diff --git a/dlls/amstream/audiostream.c b/dlls/amstream/audiostream.c
--- a/dlls/amstream/audiostream.c
+++ b/dlls/amstream/audiostream.c
@@ -960,7 +960,6 @@ static HRESULT WINAPI audio_sink_ReceiveConnection(IPin *iface, IPin *peer, con
     PIN_DIRECTION dir;
 
     TRACE("stream %p, peer %p, mt %p.\n", stream, peer, mt);
-    strmbase_dump_media_type(mt);
 
     if (!IsEqualGUID(&mt->majortype, &MEDIATYPE_Audio)
             || !IsEqualGUID(&mt->formattype, &FORMAT_WaveFormatEx)
