diff --git a/dlls/appxdeploymentclient/classes.idl b/dlls/appxdeploymentclient/classes.idl
--- a/dlls/appxdeploymentclient/classes.idl
+++ b/dlls/appxdeploymentclient/classes.idl
@@ -19,10 +19,5 @@
  */
 
 #pragma makedep register
-#pragma winrt ns_prefix
 
-import "windows.management.deployment.idl";
-
-namespace Windows.Management.Deployment {
-    runtimeclass PackageManager;
-}
+#include "windows.management.deployment.idl"
