diff --git a/dlls/cards/version.rc b/dlls/cards/version.rc
--- a/dlls/cards/version.rc
+++ b/dlls/cards/version.rc
@@ -21,5 +21,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Cards Library"
 #define WINE_FILENAME_STR "cards.dll"
 #define WINE_FILEVERSION 5,0,2134,1
+#define WINE_FILEVERSION_STR "5.00.2134.1"
+#define WINE_PRODUCTVERSION 5,0,2134,1
+#define WINE_PRODUCTVERSION_STR "5.00.2134.1"
 
 #include "wine/wine_common_ver.rc"
