diff --git a/dlls/combase/apartment.c b/dlls/combase/apartment.c
--- a/dlls/combase/apartment.c
+++ b/dlls/combase/apartment.c
@@ -719,7 +719,7 @@
         WCHAR src[MAX_PATH];
         DWORD dwLength = dstlen * sizeof(WCHAR);
 
-        if ((ret = RegQueryValueExW(regdata->u.hkey, NULL, NULL, &keytype, (BYTE*)src, &dwLength)) == ERROR_SUCCESS)
+        if ((ret = RegQueryValueExW(regdata->u.hkey, L"", NULL, &keytype, (BYTE*)src, &dwLength)) == ERROR_SUCCESS)
         {
             if (keytype == REG_EXPAND_SZ)
             {
