diff --git a/dlls/comctl32/tests/pager.c b/dlls/comctl32/tests/pager.c
--- a/dlls/comctl32/tests/pager.c
+++ b/dlls/comctl32/tests/pager.c
@@ -270,7 +270,7 @@
 static CHAR *heap_strdup(const CHAR *str)
 {
     int len = lstrlenA(str) + 1;
-    CHAR *ret = HeapAlloc(GetProcessHeap(), 0, len * sizeof(CHAR));
+    CHAR *ret = heap_alloc(len * sizeof(CHAR));
     lstrcpyA(ret, str);
     return ret;
 }
