diff --git a/dlls/browseui/browseui.rc b/dlls/browseui/browseui.rc
--- a/dlls/browseui/browseui.rc
+++ b/dlls/browseui/browseui.rc
@@ -50,6 +50,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine core dll"
 #define WINE_FILENAME_STR "browseui.dll"
 #define WINE_FILEVERSION 6,0,2900,2180
+#define WINE_FILEVERSION_STR "6.0.2900.2180"
+#define WINE_PRODUCTVERSION 6,0,2900,2180
+#define WINE_PRODUCTVERSION_STR "6.0.2900.2180"
 #define WINE_EXTRAVALUES VALUE "OLESelfRegister",""
 
 #include "wine/wine_common_ver.rc"
