diff --git a/dlls/comdlg32/comdlg32.rc b/dlls/comdlg32/comdlg32.rc
--- a/dlls/comdlg32/comdlg32.rc
+++ b/dlls/comdlg32/comdlg32.rc
@@ -535,6 +535,9 @@
 #define WINE_FILEDESCRIPTION_STR "Common Dialog Boxes"
 #define WINE_FILENAME_STR "comdlg32.dll"
 #define WINE_FILEVERSION 6,0,2900,5512
+#define WINE_FILEVERSION_STR "6.00.2900.5512"
+#define WINE_PRODUCTVERSION 6,0,2900,5512
+#define WINE_PRODUCTVERSION_STR "6.00.2900.5512"
 
 #include "wine/wine_common_ver.rc"
 
