diff --git a/dlls/gdi32/uniscribe/bracket.c b/dlls/gdi32/uniscribe/bracket.c
--- a/dlls/gdi32/uniscribe/bracket.c
+++ b/dlls/gdi32/uniscribe/bracket.c
@@ -1,5 +1,5 @@
 /* Unicode Bidirectional Bracket table */
-/* generated from https://www.unicode.org/Public/17.0.0/ucd/UCD.zip:BidiBrackets.txt */
+/* generated from https://www.unicode.org/Public/16.0.0/ucd/UCD.zip:BidiBrackets.txt */
 /* DO NOT EDIT!! */
 
 #include "windef.h"
