diff --git a/dlls/inetmib1/tests/main.c b/dlls/inetmib1/tests/main.c
--- a/dlls/inetmib1/tests/main.c
+++ b/dlls/inetmib1/tests/main.c
@@ -413,7 +413,6 @@
                         vars2[0].name.ids[vars2[0].name.idLength - 4 + i]);
                 }
             }
-            else break;
         }
         else if (noChange)
             skip("no change in OID, no MIB2 IP route table implementation\n");
