diff --git a/dlls/kernelbase/kernelbase.h b/dlls/kernelbase/kernelbase.h
--- a/dlls/kernelbase/kernelbase.h
+++ b/dlls/kernelbase/kernelbase.h
@@ -37,7 +37,6 @@
 extern void init_startup_info( RTL_USER_PROCESS_PARAMETERS *params );
 extern void init_locale( HMODULE module );
 extern void init_console(void);
-extern BOOL is_console_handle( HANDLE );
 
 extern const WCHAR windows_dir[];
 extern const WCHAR system_dir[];
