diff --git a/dlls/ntdll/tests/rtlstr.c b/dlls/ntdll/tests/rtlstr.c
--- a/dlls/ntdll/tests/rtlstr.c
+++ b/dlls/ntdll/tests/rtlstr.c
@@ -2548,7 +2548,6 @@
     truncate_expect( 8, 4, STATUS_BUFFER_TOO_SMALL);
     truncate_expect( 9, 4, STATUS_BUFFER_TOO_SMALL);
     truncate_expect(10, 5, STATUS_SUCCESS);
-    unicode_expect(special_expected, 6, 3, special_string, 7, STATUS_BUFFER_TOO_SMALL);
 #undef truncate_expect
 
     /* conversion behavior with varying input length */
