diff --git a/dlls/shell32/shellole.c b/dlls/shell32/shellole.c
--- a/dlls/shell32/shellole.c
+++ b/dlls/shell32/shellole.c
@@ -87,7 +87,6 @@
 	{&CLSID_ShellImageDataFactory, ShellImageDataFactory_Constructor},
 	{&CLSID_FileOperation, IFileOperation_Constructor},
 	{&CLSID_ActiveDesktop, ActiveDesktop_Constructor},
-	{&CLSID_EnumerableObjectCollection, EnumerableObjectCollection_Constructor},
 	{NULL, NULL}
 };
 
