diff --git a/dlls/shell32/shresdef.h b/dlls/shell32/shresdef.h
--- a/dlls/shell32/shresdef.h
+++ b/dlls/shell32/shresdef.h
@@ -227,7 +227,6 @@
 #define IDI_SHELL_BLANK3            52
 #define IDI_SHELL_BLANK4            53
 #define IDI_SHELL_UNKNOWN_DRIVE     54
-#define IDI_SHELL_SHIELD            78
 /* 55 - 132 not used on Windows */
 #define IDI_SHELL_FILES            133
 #define IDI_SHELL_FILE_SEARCH      134
