diff --git a/dlls/vulkan-1/version.rc b/dlls/vulkan-1/version.rc
--- a/dlls/vulkan-1/version.rc
+++ b/dlls/vulkan-1/version.rc
@@ -21,6 +21,7 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Vulkan Loader"
 #define WINE_FILENAME_STR "vulkan-1.dll"
 #define WINE_FILEVERSION 1,0,0,0 /* Set to 1.0.0 so the official Vulkan runtime installer will replace this file */
+#define WINE_FILEVERSION_STR PACKAGE_VERSION
 #define WINE_PRODUCTVERSION_STR PACKAGE_VERSION
 #define WINE_PRODUCTNAME_STR "Wine Vulkan"
 
