diff --git a/dlls/xmllite/xmllite_private.h b/dlls/xmllite/xmllite_private.h
--- a/dlls/xmllite/xmllite_private.h
+++ b/dlls/xmllite/xmllite_private.h
@@ -61,7 +61,6 @@
 BOOL is_ncnamechar(WCHAR ch);
 BOOL is_pubchar(WCHAR ch);
 BOOL is_namestartchar(WCHAR ch);
-BOOL is_ncnamestartchar(WCHAR ch);
 BOOL is_namechar(WCHAR ch);
 
 /* [2] Char ::= #x9 | #xA | #xD | [#x20-#xD7FF] | [#xE000-#xFFFD] | [#x10000-#x10FFFF] */
