diff --git a/dlls/crypt32/collectionstore.c b/dlls/crypt32/collectionstore.c
--- a/dlls/crypt32/collectionstore.c
+++ b/dlls/crypt32/collectionstore.c
@@ -265,7 +265,7 @@
     TRACE("(%p, %p)\n", store, cert);
 
     linked = (cert_t*)context->linked;
-    return CRYPT_DeleteCertificateFromStore(&linked->ctx);
+    return CertDeleteCertificateFromStore(&linked->ctx);
 }
 
 static BOOL Collection_addCRL(WINECRYPT_CERTSTORE *store, context_t *crl,
