diff --git a/dlls/crypt32/decode.c b/dlls/crypt32/decode.c
--- a/dlls/crypt32/decode.c
+++ b/dlls/crypt32/decode.c
@@ -6940,8 +6940,6 @@
         decodeFunc = CRYPT_AsnDecodeCRLDistPoints;
     else if (!strcmp(lpszStructType, szOID_CERT_POLICIES))
         decodeFunc = CRYPT_AsnDecodeCertPolicies;
-    else if (!strcmp(lpszStructType, szOID_APPLICATION_CERT_POLICIES))
-        FIXME("szOID_APPLICATION_CERT_POLICIES\n");
     else if (!strcmp(lpszStructType, szOID_POLICY_MAPPINGS))
         decodeFunc = CRYPT_AsnDecodeCertPolicyMappings;
     else if (!strcmp(lpszStructType, szOID_POLICY_CONSTRAINTS))
