diff --git a/dlls/d2d1/brush.c b/dlls/d2d1/brush.c
--- a/dlls/d2d1/brush.c
+++ b/dlls/d2d1/brush.c
@@ -1445,7 +1445,7 @@
             t = sqrtf(s1 - s2);
             d2d_point_set(&cb->u.radial.rb, t * -sin_theta, t * cos_theta);
 
-            cb->u.radial.stop_count = brush->u.radial.gradient->stop_count;
+            cb->u.radial.stop_count = brush->u.linear.gradient->stop_count;
 
             return TRUE;
 
