diff --git a/dlls/d3dx9_38/version.rc b/dlls/d3dx9_38/version.rc
--- a/dlls/d3dx9_38/version.rc
+++ b/dlls/d3dx9_38/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine D3DX9"
 #define WINE_FILENAME_STR "d3dx9_38.dll"
 #define WINE_FILEVERSION 9,23,949,2378
+#define WINE_FILEVERSION_STR "9.23.949.2378"
+#define WINE_PRODUCTVERSION 9,23,949,2378
+#define WINE_PRODUCTVERSION_STR "9.23.949.2378"
 
 #include "wine/wine_common_ver.rc"
