diff --git a/dlls/dbghelp/dbghelp.c b/dlls/dbghelp/dbghelp.c
--- a/dlls/dbghelp/dbghelp.c
+++ b/dlls/dbghelp/dbghelp.c
@@ -417,7 +417,7 @@
     if (!base) return FALSE;
 
     TRACE("got debug info address %#I64x from PEB %p\n", base, pbi.PebBaseAddress);
-    if (base != (ULONG_PTR)base || (!elf_read_wine_loader_dbg_info(pcs, base) && !macho_read_wine_loader_dbg_info(pcs, base)))
+    if (!elf_read_wine_loader_dbg_info(pcs, base) && !macho_read_wine_loader_dbg_info(pcs, base))
     {
         WARN("couldn't load process debug info at %#I64x\n", base);
         pcs->loader = &empty_loader_ops;
