diff --git a/dlls/dinput/tests/driver_hid.c b/dlls/dinput/tests/driver_hid.c
--- a/dlls/dinput/tests/driver_hid.c
+++ b/dlls/dinput/tests/driver_hid.c
@@ -57,8 +57,6 @@
 
     ok( device == impl->expect_hid_fdo, "got device %p\n", device );
     ok( device->DriverObject == expect_driver, "got DriverObject %p\n", device->DriverObject );
-    if (!device->NextDevice) ok( device == impl->expect_hid_fdo, "got device %p\n", device );
-    else ok( device->NextDevice == impl->expect_hid_fdo, "got NextDevice %p\n", device->NextDevice );
     ok( !device->AttachedDevice, "got AttachedDevice %p\n", device->AttachedDevice );
 
     ok( ext->MiniDeviceExtension == impl->expect_hid_ext, "got MiniDeviceExtension %p\n", ext->MiniDeviceExtension );
