diff --git a/dlls/dinput/version.rc b/dlls/dinput/version.rc
--- a/dlls/dinput/version.rc
+++ b/dlls/dinput/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine DirectInput"
 #define WINE_FILENAME_STR "dinput.dll"
 #define WINE_FILEVERSION 5,1,2600,881
+#define WINE_FILEVERSION_STR "5.1.2600.881"
+#define WINE_PRODUCTVERSION 5,1,2600,881
+#define WINE_PRODUCTVERSION_STR "5.1"
 
 #include "wine/wine_common_ver.rc"
