diff --git a/dlls/dpnhpast/version.rc b/dlls/dpnhpast/version.rc
--- a/dlls/dpnhpast/version.rc
+++ b/dlls/dpnhpast/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine DirectPlay NAT Helper PAST"
 #define WINE_FILENAME_STR "dpnhpast.dll"
 #define WINE_FILEVERSION 5,3,0,901
+#define WINE_FILEVERSION_STR "5.3.0.901"
+#define WINE_PRODUCTVERSION 5,3,0,901
+#define WINE_PRODUCTVERSION_STR "5.3"
 
 #include "wine/wine_common_ver.rc"
