diff --git a/dlls/dpnlobby/version.rc b/dlls/dpnlobby/version.rc
--- a/dlls/dpnlobby/version.rc
+++ b/dlls/dpnlobby/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine DpnLobby dll"
 #define WINE_FILENAME_STR "dpnlobby.dll"
 #define WINE_FILEVERSION 5,3,2600,2180
+#define WINE_FILEVERSION_STR "5.3.2600.2180"
+#define WINE_PRODUCTVERSION 5,3,2600,2180
+#define WINE_PRODUCTVERSION_STR "5.3.2600.2180"
 
 #include "wine/wine_common_ver.rc"
