diff --git a/dlls/dx8vb/version.rc b/dlls/dx8vb/version.rc
--- a/dlls/dx8vb/version.rc
+++ b/dlls/dx8vb/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine dx8vb"
 #define WINE_FILENAME_STR "dx8vb.dll"
 #define WINE_FILEVERSION 5,3,1,902
+#define WINE_FILEVERSION_STR "5.3.1.902"
+#define WINE_PRODUCTVERSION 5,3,1,902
+#define WINE_PRODUCTVERSION_STR "5.3.1.902"
 
 #include "wine/wine_common_ver.rc"
