diff --git a/dlls/dxdiagn/dxdiagn.rc b/dlls/dxdiagn/dxdiagn.rc
--- a/dlls/dxdiagn/dxdiagn.rc
+++ b/dlls/dxdiagn/dxdiagn.rc
@@ -33,5 +33,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine DxDiag 8"
 #define WINE_FILENAME_STR "dxdiagn.dll"
 #define WINE_FILEVERSION 5,3,1,904
+#define WINE_FILEVERSION_STR "5.3.1.904"
+#define WINE_PRODUCTVERSION 5,3,1,904
+#define WINE_PRODUCTVERSION_STR "5.3.1.904"
 
 #include "wine/wine_common_ver.rc"
