diff --git a/dlls/kernel32/Makefile.in b/dlls/kernel32/Makefile.in
--- a/dlls/kernel32/Makefile.in
+++ b/dlls/kernel32/Makefile.in
@@ -1,7 +1,7 @@
 EXTRADEFS = -D_KERNEL32_ -D_NORMALIZE_
 MODULE    = kernel32.dll
 IMPORTLIB = kernel32
-IMPORTS   = kernelbase ntdll
+IMPORTS   = kernelbase ntdll winecrt0
 
 EXTRADLLFLAGS = -nodefaultlibs -Wb,-F,KERNEL32.dll
 i386_EXTRADLLFLAGS = -Wl,--image-base,0x7b800000
