diff --git a/dlls/kernelbase/Makefile.in b/dlls/kernelbase/Makefile.in
--- a/dlls/kernelbase/Makefile.in
+++ b/dlls/kernelbase/Makefile.in
@@ -1,7 +1,7 @@
 EXTRADEFS = -DWINBASEAPI= -DWINADVAPI= -DWINUSERAPI= -DWINSHLWAPI=
 MODULE    = kernelbase.dll
 IMPORTLIB = kernelbase
-IMPORTS   = uuid ntdll
+IMPORTS   = uuid ntdll winecrt0
 
 EXTRADLLFLAGS = -nodefaultlibs -nostartfiles
 i386_EXTRADLLFLAGS = -Wl,--image-base,0x7b000000
