diff --git a/dlls/krnl386.exe16/fpu.c b/dlls/krnl386.exe16/fpu.c
--- a/dlls/krnl386.exe16/fpu.c
+++ b/dlls/krnl386.exe16/fpu.c
@@ -221,7 +221,7 @@
  */
 static void FPU_ModifyCode(CONTEXT *context, BYTE Opcode)
 {
-    BYTE *code = ldt_get_ptr(context->SegCs, context->Eip);
+    BYTE *code = CTX_SEG_OFF_TO_LIN(context, context->SegCs, context->Eip);
 
     /*
      * All *NIX systems should have a real or kernel emulated FPU.
