diff --git a/dlls/mountmgr.sys/cred.c b/dlls/mountmgr.sys/cred.c
--- a/dlls/mountmgr.sys/cred.c
+++ b/dlls/mountmgr.sys/cred.c
@@ -62,7 +62,7 @@
 
     if (!(query_cfstring = CFStringCreateWithFormat(kCFAllocatorDefault, NULL,
                                                     CFSTR("com_apple_xcode_dsym_uuids == \"%08X-%04X-%04X-%02X%02X-%02X%02X%02X%02X%02X%02X\""),
-                                                    id->Data1, id->Data2, id->Data3, id->Data4[0],
+                                                    (unsigned int)id->Data1, id->Data2, id->Data3, id->Data4[0],
                                                     id->Data4[1], id->Data4[2], id->Data4[3], id->Data4[4],
                                                     id->Data4[5], id->Data4[6], id->Data4[7] )))
         return STATUS_NO_MEMORY;
