diff --git a/dlls/mpr/mpr.rc b/dlls/mpr/mpr.rc
--- a/dlls/mpr/mpr.rc
+++ b/dlls/mpr/mpr.rc
@@ -54,5 +54,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Multiprotocol Router Library"
 #define WINE_FILENAME_STR "mpr.dll"
 #define WINE_FILEVERSION 5,0,2195,6611
+#define WINE_FILEVERSION_STR "5.00.2195.6611"
+#define WINE_PRODUCTVERSION 5,0,2195,6611
+#define WINE_PRODUCTVERSION_STR "5.00.2195.6611"
 
 #include "wine/wine_common_ver.rc"
