diff --git a/dlls/msctf/version.rc b/dlls/msctf/version.rc
--- a/dlls/msctf/version.rc
+++ b/dlls/msctf/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Msctf"
 #define WINE_FILENAME_STR "msctf.dll"
 #define WINE_FILEVERSION 5,1,2600,3319
+#define WINE_FILEVERSION_STR " 5.1.2600.3319"
+#define WINE_PRODUCTVERSION 5,1,2600,3319
+#define WINE_PRODUCTVERSION_STR "5.1.2600.3319"
 
 #include "wine/wine_common_ver.rc"
