diff --git a/dlls/mshtml/selection.c b/dlls/mshtml/selection.c
--- a/dlls/mshtml/selection.c
+++ b/dlls/mshtml/selection.c
@@ -239,7 +239,7 @@
     0
 };
 dispex_static_data_t MSSelection_dispex = {
-    .id              = OBJID_MSSelection,
+    .id              = PROT_MSSelection,
     .vtbl            = &HTMLSelectionObject_dispex_vtbl,
     .disp_tid        = IHTMLSelectionObject_tid,
     .iface_tids      = MSSelection_iface_tids,
