diff --git a/dlls/mshtml/tests/xmlhttprequest.c b/dlls/mshtml/tests/xmlhttprequest.c
--- a/dlls/mshtml/tests/xmlhttprequest.c
+++ b/dlls/mshtml/tests/xmlhttprequest.c
@@ -1233,7 +1233,7 @@
     CHECK_CALLED(xdomainrequest_onload);
 
     hres = IHTMLXDomainRequest_get_responseText(xdr, &bstr);
-    ok(hres == S_OK, "get_responseText returned %08lx\n", hres);
+    ok(hres == S_OK, "get_contentType returned %08lx\n", hres);
     ok(!lstrcmpW(bstr, L"<html><body>test</body></html>\n"), "responseText = %s\n", debugstr_w(bstr));
     SysFreeString(bstr);
 
