diff --git a/dlls/mspatcha/version.rc b/dlls/mspatcha/version.rc
--- a/dlls/mspatcha/version.rc
+++ b/dlls/mspatcha/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine File Patch API"
 #define WINE_FILENAME_STR "mspatcha.dll"
 #define WINE_FILEVERSION 6,1,7600,16385
+#define WINE_FILEVERSION_STR "6.1.7600.16385"
+#define WINE_PRODUCTVERSION 6,1,7600,16385
+#define WINE_PRODUCTVERSION_STR "6.1.7600.16385"
 
 #include "wine/wine_common_ver.rc"
