diff --git a/dlls/msvcp60/main.c b/dlls/msvcp60/main.c
--- a/dlls/msvcp60/main.c
+++ b/dlls/msvcp60/main.c
@@ -58,7 +58,7 @@
 /* ?_Fpz@std@@3_JB */
 const __int64 std_Fpz = 0;
 
-static void* (__cdecl __WINE_ALLOC_SIZE(1) *MSVCRT_operator_new)(size_t);
+static void* (__cdecl *MSVCRT_operator_new)(size_t);
 static void (__cdecl *MSVCRT_operator_delete)(void*);
 void* (__cdecl *MSVCRT_set_new_handler)(void*);
 
