diff --git a/dlls/msvcp90/msvcp_main.c b/dlls/msvcp90/msvcp_main.c
--- a/dlls/msvcp90/msvcp_main.c
+++ b/dlls/msvcp90/msvcp_main.c
@@ -246,7 +246,7 @@
    _invalid_parameter_noinfo();
 }
 #else
-static void* (__cdecl __WINE_ALLOC_SIZE(1) *MSVCRT_operator_new)(size_t);
+static void* (__cdecl *MSVCRT_operator_new)(size_t);
 static void (__cdecl *MSVCRT_operator_delete)(void*);
 
 void* __cdecl operator_new(size_t size)
