diff --git a/dlls/msvcrt/tests/data.c b/dlls/msvcrt/tests/data.c
--- a/dlls/msvcrt/tests/data.c
+++ b/dlls/msvcrt/tests/data.c
@@ -217,7 +217,7 @@
     memset(&startup, 0, sizeof(startup));
     startup.cb = sizeof(startup);
     CreateProcessA(NULL, cmdline, NULL, NULL, TRUE, CREATE_DEFAULT_ERROR_MODE|NORMAL_PRIORITY_CLASS, NULL, NULL, &startup, &proc);
-    wait_child_process(&proc);
+    wait_child_process(proc.hProcess);
 
     _unlink(filepath);
     sprintf(filepath, "%swine_test\\a", tmppath);
