diff --git a/dlls/nsiproxy.sys/ip.c b/dlls/nsiproxy.sys/ip.c
--- a/dlls/nsiproxy.sys/ip.c
+++ b/dlls/nsiproxy.sys/ip.c
@@ -1517,7 +1517,7 @@
             {
                 entry.prefix.s_addr = htonl( 0x7f000000 );
                 entry.next_hop.s_addr = 0;
-                entry.metric = ~0u;
+                entry.metric = 256;
                 entry.prefix_len = 8;
                 entry.protocol = MIB_IPPROTO_LOCAL;
                 entry.loopback = 1;
