diff --git a/dlls/odbc32/rsrc.rc b/dlls/odbc32/rsrc.rc
--- a/dlls/odbc32/rsrc.rc
+++ b/dlls/odbc32/rsrc.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine ODBC Library"
 #define WINE_FILENAME_STR "odbc32.dll"
 #define WINE_FILEVERSION 3,520,6301,0
+#define WINE_FILEVERSION_STR "3.520.6301.0"
+#define WINE_PRODUCTVERSION 3,520,6301,0
+#define WINE_PRODUCTVERSION_STR "3.520.6301.0"
 
 #include "wine/wine_common_ver.rc"
