diff --git a/dlls/ole32/storage32.c b/dlls/ole32/storage32.c
--- a/dlls/ole32/storage32.c
+++ b/dlls/ole32/storage32.c
@@ -2895,7 +2895,7 @@
   offset.LowPart = 0;
   hr = StorageImpl_ReadAt(This, offset, headerBigBlock, HEADER_SIZE, &bytes_read);
   if (SUCCEEDED(hr) && bytes_read != HEADER_SIZE)
-    hr = STG_E_FILEALREADYEXISTS;
+    hr = STG_E_FILENOTFOUND;
 
   /*
    * Extract the information from the header.
