diff --git a/dlls/oleacc/tests/main.c b/dlls/oleacc/tests/main.c
--- a/dlls/oleacc/tests/main.c
+++ b/dlls/oleacc/tests/main.c
@@ -727,7 +727,7 @@
     memset(&startup, 0, sizeof(startup));
     startup.cb = sizeof(startup);
     CreateProcessA(NULL, cmdline, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &proc);
-    wait_child_process(&proc);
+    wait_child_process(proc.hProcess);
     ok(Object_ref == 1, "Object_ref = %ld\n", Object_ref);
 }
 
