diff --git a/dlls/opcservices/uri.c b/dlls/opcservices/uri.c
--- a/dlls/opcservices/uri.c
+++ b/dlls/opcservices/uri.c
@@ -477,7 +477,7 @@
     {
         if (!wcscmp(path + len - ARRAY_SIZE(relsextW) + 1, relsextW))
         {
-            if (start && !memcmp(start, relsdirW, sizeof(relsdirW) - sizeof(WCHAR)))
+            if (start && !memcmp(start, relsdirW, ARRAY_SIZE(relsdirW) - sizeof(WCHAR)))
             {
                 SysFreeString(path);
                 return NULL;
