diff --git a/dlls/psapi/version.rc b/dlls/psapi/version.rc
--- a/dlls/psapi/version.rc
+++ b/dlls/psapi/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Process Status Helper"
 #define WINE_FILENAME_STR "psapi.dll"
 #define WINE_FILEVERSION 5,1,2600,3264
+#define WINE_FILEVERSION_STR "5.1.2600.3264"
+#define WINE_PRODUCTVERSION 5,1,2600,3264
+#define WINE_PRODUCTVERSION_STR "5.1.2600.3264"
 
 #include "wine/wine_common_ver.rc"
