diff --git a/dlls/riched32/version.rc b/dlls/riched32/version.rc
--- a/dlls/riched32/version.rc
+++ b/dlls/riched32/version.rc
@@ -17,5 +17,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wrapper Dll for Richedit"
 #define WINE_FILENAME_STR "riched32.dll"
 #define WINE_FILEVERSION 5,1,2600,0
+#define WINE_FILEVERSION_STR "5,1,2600,0"
+#define WINE_PRODUCTVERSION 5,1,2600,0
+#define WINE_PRODUCTVERSION_STR "5,1,2600,0"
 
 #include "wine/wine_common_ver.rc"
