diff --git a/dlls/sechost/sechost.spec b/dlls/sechost/sechost.spec
--- a/dlls/sechost/sechost.spec
+++ b/dlls/sechost/sechost.spec
@@ -89,7 +89,7 @@
 @ stdcall EnumDependentServicesW(long long ptr long ptr ptr)
 @ stdcall EnumServicesStatusExW(long long long long ptr long ptr ptr ptr wstr)
 @ stub EnumerateIdentityProviders
-@ stdcall EnumerateTraceGuidsEx(long ptr long ptr long ptr) advapi32.EnumerateTraceGuidsEx
+@ stub EnumerateTraceGuidsEx
 @ stub EtwQueryRealtimeConsumer
 @ stub EventAccessControl
 @ stub EventAccessQuery
