diff --git a/dlls/shell.dll16/shell.c b/dlls/shell.dll16/shell.c
--- a/dlls/shell.dll16/shell.c
+++ b/dlls/shell.dll16/shell.c
@@ -234,7 +234,7 @@
  */
 BOOL16 WINAPI ShellAbout16( HWND16 hWnd, LPCSTR szApp, LPCSTR szOtherStuff, HICON16 icon16 )
 {
-    HICON icon = icon16 ? convert_icon_to_32( icon16 ) : 0;
+    HICON icon = convert_icon_to_32( icon16 );
     BOOL ret = ShellAboutA( HWND_32(hWnd), szApp, szOtherStuff, icon );
     DestroyIcon( icon );
     return ret;
