diff --git a/dlls/shlwapi/shlwapi.rc b/dlls/shlwapi/shlwapi.rc
--- a/dlls/shlwapi/shlwapi.rc
+++ b/dlls/shlwapi/shlwapi.rc
@@ -50,5 +50,8 @@
 
 #define WINE_FILENAME_STR "shlwapi.dll"
 #define WINE_FILEVERSION 6,0,2800,1692
+#define WINE_FILEVERSION_STR "6.0.2800.1692"
+#define WINE_PRODUCTVERSION 6,0,2800,1692
+#define WINE_PRODUCTVERSION_STR "6.0.2800.1692"
 
 #include "wine/wine_common_ver.rc"
