diff --git a/dlls/twinapi.appcore/main.c b/dlls/twinapi.appcore/main.c
--- a/dlls/twinapi.appcore/main.c
+++ b/dlls/twinapi.appcore/main.c
@@ -75,8 +75,6 @@
         IActivationFactory_QueryInterface( analytics_info_factory, &IID_IActivationFactory, (void **)factory );
     else if (!wcscmp( buffer, RuntimeClass_Windows_System_UserProfile_AdvertisingManager ))
         IActivationFactory_QueryInterface( advertising_manager_factory, &IID_IActivationFactory, (void **)factory );
-    else if (!wcscmp( buffer, RuntimeClass_Windows_UI_ViewManagement_ApplicationView ))
-        IActivationFactory_QueryInterface( application_view_factory, &IID_IActivationFactory, (void **)factory );
 
     if (*factory) return S_OK;
     return CLASS_E_CLASSNOTAVAILABLE;
