diff --git a/dlls/ucrtbase/version.rc b/dlls/ucrtbase/version.rc
--- a/dlls/ucrtbase/version.rc
+++ b/dlls/ucrtbase/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine runtime library"
 #define WINE_FILENAME_STR "ucrtbase.dll"
 #define WINE_FILEVERSION 10,0,14393,2247
+#define WINE_FILEVERSION_STR "10.0.14393.2247"
+#define WINE_PRODUCTVERSION 10,0,14393,2247
+#define WINE_PRODUCTVERSION_STR "10.0.14393.2247"
 
 #include "wine/wine_common_ver.rc"
