diff --git a/dlls/usp10/tests/usp10.c b/dlls/usp10/tests/usp10.c
--- a/dlls/usp10/tests/usp10.c
+++ b/dlls/usp10/tests/usp10.c
@@ -1030,7 +1030,7 @@
         winetest_ok(nGlyphs == outnGlyphs, "got incorrect number of glyphs (%i)\n",outnGlyphs);
     else if (nGlyphs != outnGlyphs)
         winetest_trace("got incorrect number of glyphs (%i)\n",outnGlyphs);
-    for (x = 0; x < min(nGlyphs, outnGlyphs); x++)
+    for (x = 0; x < outnGlyphs; x++)
     {
         if (glyphItems[x].Glyph)
         {
