diff --git a/dlls/wbemprox/reg.c b/dlls/wbemprox/reg.c
--- a/dlls/wbemprox/reg.c
+++ b/dlls/wbemprox/reg.c
@@ -106,7 +106,7 @@
 
     TRACE("%p, %s\n", root, debugstr_w(subkey));
 
-    res = RegCreateKeyExW( root, subkey, 0, NULL, 0, KEY_ENUMERATE_SUB_KEYS | get_access_mask( context ), NULL, &hkey, NULL );
+    res = RegCreateKeyExW( root, subkey, 0, NULL, 0, get_access_mask( context ), NULL, &hkey, NULL );
     set_variant( VT_UI4, res, NULL, retval );
     if (!res)
     {
