diff --git a/dlls/windows.applicationmodel/main.c b/dlls/windows.applicationmodel/main.c
--- a/dlls/windows.applicationmodel/main.c
+++ b/dlls/windows.applicationmodel/main.c
@@ -40,8 +40,6 @@
 
     if (!wcscmp( buffer, RuntimeClass_Windows_ApplicationModel_Package ))
         IActivationFactory_QueryInterface( package_factory, &IID_IActivationFactory, (void **)factory );
-    else if (!wcscmp( buffer, RuntimeClass_Windows_ApplicationModel_DesignMode ))
-        IActivationFactory_QueryInterface( design_mode_factory, &IID_IActivationFactory, (void **)factory );
 
     if (*factory) return S_OK;
     return CLASS_E_CLASSNOTAVAILABLE;
