diff --git a/dlls/windows.security.credentials.ui.userconsentverifier/Makefile.in b/dlls/windows.security.credentials.ui.userconsentverifier/Makefile.in
--- a/dlls/windows.security.credentials.ui.userconsentverifier/Makefile.in
+++ b/dlls/windows.security.credentials.ui.userconsentverifier/Makefile.in
@@ -3,6 +3,6 @@
 
 SOURCES = \
 	async.c \
-	async_private.idl \
 	classes.idl \
-	main.c
+	main.c \
+	provider.idl
