diff --git a/dlls/windowscodecs/encoder.c b/dlls/windowscodecs/encoder.c
--- a/dlls/windowscodecs/encoder.c
+++ b/dlls/windowscodecs/encoder.c
@@ -479,7 +479,7 @@
     if (!(encoder->parent->encoder_info.flags & ENCODER_FLAGS_SUPPORTS_METADATA))
         return WINCODEC_ERR_UNSUPPORTEDOPERATION;
 
-    return MetadataQueryWriter_CreateInstanceFromBlockWriter(&encoder->IWICMetadataBlockWriter_iface, ppIMetadataQueryWriter);
+    return MetadataQueryWriter_CreateInstance(&encoder->IWICMetadataBlockWriter_iface, NULL, ppIMetadataQueryWriter);
 }
 
 static const IWICBitmapFrameEncodeVtbl CommonEncoderFrame_Vtbl = {
