diff --git a/dlls/wininet/wininet.rc b/dlls/wininet/wininet.rc
--- a/dlls/wininet/wininet.rc
+++ b/dlls/wininet/wininet.rc
@@ -99,5 +99,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Internet Connectivity"
 #define WINE_FILENAME_STR "wininet.dll"
 #define WINE_FILEVERSION 8,00,7601,17601
+#define WINE_FILEVERSION_STR "8.00.7601.17601"
+#define WINE_PRODUCTVERSION 8,00,7601,17601
+#define WINE_PRODUCTVERSION_STR "8.00.7601.17601"
 
 #include "wine/wine_common_ver.rc"
