diff --git a/dlls/wintrust/version.rc b/dlls/wintrust/version.rc
--- a/dlls/wintrust/version.rc
+++ b/dlls/wintrust/version.rc
@@ -20,5 +20,8 @@
 
 #define WINE_FILENAME_STR "wintrust.dll"
 #define WINE_FILEVERSION 6,1,7601,23769
+#define WINE_FILEVERSION_STR "6.1.7601.23769"
+#define WINE_PRODUCTVERSION 6,1,7601,23769
+#define WINE_PRODUCTVERSION_STR "6.1.7601.23769"
 
 #include "wine/wine_common_ver.rc"
