diff --git a/dlls/wmvcore/version.rc b/dlls/wmvcore/version.rc
--- a/dlls/wmvcore/version.rc
+++ b/dlls/wmvcore/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Media Playback dll"
 #define WINE_FILENAME_STR "wmvcore.dll"
 #define WINE_FILEVERSION 12,0,7601,17514
+#define WINE_FILEVERSION_STR "12.0.7601.17514"
+#define WINE_PRODUCTVERSION 12,0,7601,17514
+#define WINE_PRODUCTVERSION_STR "12.0.7601.17514"
 
 #include "wine/wine_common_ver.rc"
