diff --git a/dlls/wpcap/version.rc b/dlls/wpcap/version.rc
--- a/dlls/wpcap/version.rc
+++ b/dlls/wpcap/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "wpcap.dll Dynamic Link Library - based on libpcap 1.0"
 #define WINE_FILENAME_STR "wpcap.dll"
 #define WINE_FILEVERSION 4,1,0,2980
+#define WINE_FILEVERSION_STR "4.1.0.2980"
+#define WINE_PRODUCTVERSION 4,1,0,2980
+#define WINE_PRODUCTVERSION_STR "4.1.0.2980"
 
 #include "wine/wine_common_ver.rc"
