diff --git a/dlls/ws2_32/version.rc b/dlls/ws2_32/version.rc
--- a/dlls/ws2_32/version.rc
+++ b/dlls/ws2_32/version.rc
@@ -31,5 +31,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine core dll"
 #define WINE_FILENAME_STR "ws2_32.dll"
 #define WINE_FILEVERSION 5,1,2600,5512
+#define WINE_FILEVERSION_STR "5.1.2600.5512"
+#define WINE_PRODUCTVERSION 5,1,2600,5512
+#define WINE_PRODUCTVERSION_STR "5.1.2600.5512"
 
 #include "wine/wine_common_ver.rc"
