diff --git a/dlls/xaudio2_9/version.rc b/dlls/xaudio2_9/version.rc
--- a/dlls/xaudio2_9/version.rc
+++ b/dlls/xaudio2_9/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine xaudio2_9"
 #define WINE_FILENAME_STR "xaudio2_9.dll"
 #define WINE_FILEVERSION 10,0,19041,1
+#define WINE_FILEVERSION_STR "10.0.19041.1"
+#define WINE_PRODUCTVERSION 10,0,19041,1
+#define WINE_PRODUCTVERSION_STR "10.0.19041.1"
 
 #include "wine/wine_common_ver.rc"
