diff --git a/dlls/xinputuap/version.rc b/dlls/xinputuap/version.rc
--- a/dlls/xinputuap/version.rc
+++ b/dlls/xinputuap/version.rc
@@ -20,5 +20,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Common Controller API"
 #define WINE_FILENAME_STR "xinputuap.dll"
 #define WINE_FILEVERSION 10,0,19041,746
+#define WINE_FILEVERSION_STR "10.0.19041.746"
+#define WINE_PRODUCTVERSION 10,0,19041,746
+#define WINE_PRODUCTVERSION_STR "10.0"
 
 #include "wine/wine_common_ver.rc"
