diff --git a/dlls/xpsprint/xpsprint.rc b/dlls/xpsprint/xpsprint.rc
--- a/dlls/xpsprint/xpsprint.rc
+++ b/dlls/xpsprint/xpsprint.rc
@@ -21,5 +21,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine XPS Print dll"
 #define WINE_FILENAME_STR "xpsprint.dll"
 #define WINE_FILEVERSION 7,0,6002,18363
+#define WINE_FILEVERSION_STR "7.0.6002.18363"
+#define WINE_PRODUCTVERSION 7,0,6002,18363
+#define WINE_PRODUCTVERSION_STR "7.0.6002.18363"
 
 #include "wine/wine_common_ver.rc"
