diff --git a/dlls/concrt140/details.h b/dlls/concrt140/details.h
--- a/dlls/concrt140/details.h
+++ b/dlls/concrt140/details.h
@@ -22,8 +22,8 @@
 #include "winternl.h"
 #include "cppexcept.h"
 
+void* __cdecl operator_new(size_t);
 void __cdecl operator_delete(void*);
-void* __cdecl operator_new(size_t) __WINE_ALLOC_SIZE(1) __WINE_DEALLOC(operator_delete) __WINE_MALLOC;
 
 bool __cdecl Context_IsCurrentTaskCollectionCanceling(void);
 
