diff --git a/dlls/d2d1/d2d1.spec b/dlls/d2d1/d2d1.spec
--- a/dlls/d2d1/d2d1.spec
+++ b/dlls/d2d1/d2d1.spec
@@ -5,8 +5,7 @@
 @ stdcall D2D1InvertMatrix(ptr)
 @ stdcall D2D1ConvertColorSpace(long long ptr)
 @ stdcall D2D1CreateDevice(ptr ptr ptr)
-@ stdcall D2D1CreateDeviceContext(ptr ptr ptr)
+@ stub D2D1CreateDeviceContext
 @ stdcall D2D1SinCos(float ptr ptr)
 @ stdcall D2D1Tan(float)
 @ stdcall D2D1Vec3Length(float float float)
-@ stdcall D2D1ComputeMaximumScaleFactor(ptr)
