diff --git a/dlls/d3d11/d3d11_main.c b/dlls/d3d11/d3d11_main.c
--- a/dlls/d3d11/d3d11_main.c
+++ b/dlls/d3d11/d3d11_main.c
@@ -115,8 +115,7 @@
     TRACE("factory %p, adapter %p, flags %#x, feature_levels %p, levels %u, device %p.\n",
             factory, adapter, flags, feature_levels, levels, device);
 
-    GetModuleHandleExW(GET_MODULE_HANDLE_EX_FLAG_FROM_ADDRESS
-            | GET_MODULE_HANDLE_EX_FLAG_UNCHANGED_REFCOUNT, (const WCHAR *)D3D11CoreCreateDevice, &d3d11);
+    d3d11 = GetModuleHandleA("d3d11.dll");
     hr = DXGID3D10CreateDevice(d3d11, factory, adapter, flags, feature_levels, levels, (void **)&dxgi_device);
     if (FAILED(hr))
     {
