diff --git a/dlls/d3dx9_43/Makefile.in b/dlls/d3dx9_43/Makefile.in
--- a/dlls/d3dx9_43/Makefile.in
+++ b/dlls/d3dx9_43/Makefile.in
@@ -1,7 +1,7 @@
-EXTRADEFS = -DD3DX_SDK_VERSION=43
+EXTRADEFS = -DD3DX_SDK_VERSION=43 -DD3DX_D3D_VERSION=9
 MODULE    = d3dx9_43.dll
 IMPORTLIB = d3dx9_43
-IMPORTS   = d3d9 d3dcompiler_43 dxguid d3dxof ole32 gdi32 user32
+IMPORTS   = d3d9 d3dcompiler dxguid d3dxof ole32 gdi32 user32
 PARENTSRC = ../d3dx9_36
 DELAYIMPORTS = windowscodecs
 
