diff --git a/dlls/dsound/version.rc b/dlls/dsound/version.rc
--- a/dlls/dsound/version.rc
+++ b/dlls/dsound/version.rc
@@ -19,6 +19,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine DirectSound"
 #define WINE_FILENAME_STR "dsound.dll"
 #define WINE_FILEVERSION 5,3,1,904
+#define WINE_FILEVERSION_STR "5.3.1.904"
+#define WINE_PRODUCTVERSION 5,3,1,904
+#define WINE_PRODUCTVERSION_STR "5.3.1.904"
 #define WINE_EXTRAVALUES VALUE "OLESelfRegister",""
 
 #include "wine/wine_common_ver.rc"
