diff --git a/dlls/explorerframe/version.rc b/dlls/explorerframe/version.rc
--- a/dlls/explorerframe/version.rc
+++ b/dlls/explorerframe/version.rc
@@ -21,6 +21,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine ExplorerFrame"
 #define WINE_FILENAME_STR "explorerframe.dll"
 #define WINE_FILEVERSION 6,1,7601,16666
+#define WINE_FILEVERSION_STR "6.1.7601.16666"
+#define WINE_PRODUCTVERSION 6,1,7601,16666
+#define WINE_PRODUCTVERSION_STR "6.1.7601.16666"
 #define WINE_EXTRAVALUES VALUE "OLESelfRegister",""
 
 #include "wine/wine_common_ver.rc"
