diff --git a/dlls/gdiplus/gdiplus.rc b/dlls/gdiplus/gdiplus.rc
--- a/dlls/gdiplus/gdiplus.rc
+++ b/dlls/gdiplus/gdiplus.rc
@@ -21,6 +21,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine gdiplus"
 #define WINE_FILENAME_STR "gdiplus.dll"
 #define WINE_FILEVERSION 6,1,7601,17514
+#define WINE_FILEVERSION_STR "6.1.7601.17514"
+#define WINE_PRODUCTVERSION 6,1,7601,17514
+#define WINE_PRODUCTVERSION_STR "6.1.7601.17514"
 
 #include "wine/wine_common_ver.rc"
 
