diff --git a/dlls/ieframe/ieframe.rc b/dlls/ieframe/ieframe.rc
--- a/dlls/ieframe/ieframe.rc
+++ b/dlls/ieframe/ieframe.rc
@@ -101,6 +101,9 @@
 #define WINE_FILEDESCRIPTION_STR "Internet Browser"
 #define WINE_FILENAME_STR "ieframe.dll"
 #define WINE_FILEVERSION 8,00,7601,17601
+#define WINE_FILEVERSION_STR "8.00.7601.17601"
+#define WINE_PRODUCTVERSION 8,00,7601,17601
+#define WINE_PRODUCTVERSION_STR "8.00.7601.17601"
 #define WINE_EXTRAVALUES VALUE "OLESelfRegister",""
 
 #include "wine/wine_common_ver.rc"
