diff --git a/dlls/krnl386.exe16/kernel.c b/dlls/krnl386.exe16/kernel.c
--- a/dlls/krnl386.exe16/kernel.c
+++ b/dlls/krnl386.exe16/kernel.c
@@ -129,7 +129,8 @@
     /* Initialize the real-mode selector entry points */
 #define SET_ENTRY_POINT( num, addr ) \
     NE_SetEntryPoint( inst, (num), GLOBAL_CreateBlock( GMEM_FIXED, \
-                      DOSMEM_MapDosToLinear(addr), 0x10000, inst, data_segment ))
+                      DOSMEM_MapDosToLinear(addr), 0x10000, inst, \
+                      LDT_FLAGS_DATA ))
 
     SET_ENTRY_POINT( 174, 0xa0000 );  /* KERNEL.174: __A000H */
     SET_ENTRY_POINT( 181, 0xb0000 );  /* KERNEL.181: __B000H */
