diff --git a/dlls/mshtml/mshtml.rc b/dlls/mshtml/mshtml.rc
--- a/dlls/mshtml/mshtml.rc
+++ b/dlls/mshtml/mshtml.rc
@@ -62,6 +62,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine HTML Viewer"
 #define WINE_FILENAME_STR "mshtml.dll"
 #define WINE_FILEVERSION 6,0,2800,1106
+#define WINE_FILEVERSION_STR "6.0.2800.1106"
+#define WINE_PRODUCTVERSION 6,0,2800,1106
+#define WINE_PRODUCTVERSION_STR "6.0.2800.1106"
 #define WINE_EXTRAVALUES VALUE "OLESelfRegister",""
 
 #include "wine/wine_common_ver.rc"
