diff --git a/dlls/ntdll/process.c b/dlls/ntdll/process.c
--- a/dlls/ntdll/process.c
+++ b/dlls/ntdll/process.c
@@ -265,9 +265,6 @@
 
     sel = info->Selector | 3;
     if (sel == 0x03) goto done; /* null selector */
-
-    if (sel & 0x04) /* LDT selector */
-        return NtQueryInformationThread( handle, ThreadDescriptorTableEntry, info, size, NULL );
 
     /* set common data */
     entry.HighWord.Bits.Dpl = 3;
