diff --git a/dlls/oledb32/version.rc b/dlls/oledb32/version.rc
--- a/dlls/oledb32/version.rc
+++ b/dlls/oledb32/version.rc
@@ -23,6 +23,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine oledb"
 #define WINE_FILENAME_STR "oledb32.dll"
 #define WINE_FILEVERSION 2,81,1117,0
+#define WINE_FILEVERSION_STR "2.81.1117.0"
+#define WINE_PRODUCTVERSION 2,81,1117,0
+#define WINE_PRODUCTVERSION_STR "2.81.1117.0"
 
 #include "wine/wine_common_ver.rc"
 
