diff --git a/dlls/photometadatahandler/photometadatahandler.rc b/dlls/photometadatahandler/photometadatahandler.rc
--- a/dlls/photometadatahandler/photometadatahandler.rc
+++ b/dlls/photometadatahandler/photometadatahandler.rc
@@ -21,5 +21,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine photometadatahandler dll"
 #define WINE_FILENAME_STR "photometadatahandler.dll"
 #define WINE_FILEVERSION 7,0,0,0
+#define WINE_FILEVERSION_STR "7.0.0.0"
+#define WINE_PRODUCTVERSION 7,0,0,0
+#define WINE_PRODUCTVERSION_STR "7.0.0.0"
+
 
 #include "wine/wine_common_ver.rc"
