diff --git a/dlls/riched20/version.rc b/dlls/riched20/version.rc
--- a/dlls/riched20/version.rc
+++ b/dlls/riched20/version.rc
@@ -19,6 +19,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Richedit dll"
 #define WINE_FILENAME_STR "riched20.dll"
 #define WINE_FILEVERSION 5,30,23,1215
+#define WINE_FILEVERSION_STR "5,30,23,1215"
+#define WINE_PRODUCTVERSION 5,30,23,1215
+#define WINE_PRODUCTVERSION_STR "5,30,23,1215"
 
 #include "wine/wine_common_ver.rc"
 
