diff --git a/dlls/user32/controls.h b/dlls/user32/controls.h
--- a/dlls/user32/controls.h
+++ b/dlls/user32/controls.h
@@ -78,6 +78,9 @@
 extern LRESULT ScrollBarWndProc_common(HWND,UINT,WPARAM,LPARAM,BOOL);
 extern LRESULT StaticWndProc_common(HWND,UINT,WPARAM,LPARAM,BOOL);
 
+/* Class functions */
+extern ATOM get_int_atom_value( UNICODE_STRING *name );
+
 /* desktop */
 extern BOOL update_wallpaper( const WCHAR *wallpaper, const WCHAR *pattern );
 
