diff --git a/dlls/winegstreamer/color_convert.c b/dlls/winegstreamer/color_convert.c
--- a/dlls/winegstreamer/color_convert.c
+++ b/dlls/winegstreamer/color_convert.c
@@ -665,7 +665,8 @@
     if (!samples->pSample)
         return E_INVALIDARG;
 
-    if (SUCCEEDED(hr = wg_transform_read_mf(impl->wg_transform, samples->pSample, &samples->dwStatus, NULL)))
+    if (SUCCEEDED(hr = wg_transform_read_mf(impl->wg_transform, samples->pSample,
+            impl->output_info.cbSize, &samples->dwStatus, NULL)))
         wg_sample_queue_flush(impl->wg_sample_queue, false);
 
     return hr;
