diff --git a/dlls/dmloader/loaderstream.c b/dlls/dmloader/loaderstream.c
--- a/dlls/dmloader/loaderstream.c
+++ b/dlls/dmloader/loaderstream.c
@@ -407,8 +407,8 @@
     if (SUCCEEDED(hr = file_stream_create(This->path, ret_iface)))
     {
         LARGE_INTEGER position = {0};
-        position.LowPart = SetFilePointer(This->file, 0, NULL, FILE_CURRENT);
-        hr = IStream_Seek(*ret_iface, position, STREAM_SEEK_SET, NULL);
+        position.LowPart = SetFilePointer(This->file, 0, NULL, SEEK_CUR);
+        hr = IStream_Seek(*ret_iface, position, SEEK_SET, NULL);
     }
 
     return hr;
