diff --git a/dlls/dmstyle/dmutils.h b/dlls/dmstyle/dmutils.h
--- a/dlls/dmstyle/dmutils.h
+++ b/dlls/dmstyle/dmutils.h
@@ -27,4 +27,10 @@
 	DWORD dwSize; /* size of the chunk */
 } DMUS_PRIVATE_CHUNK, *LPDMUS_PRIVATE_CHUNK;
 
+/**
+ * Parsing utilities
+ */
+extern HRESULT IDirectMusicUtils_IPersistStream_ParseDescGeneric (DMUS_PRIVATE_CHUNK* pChunk, IStream* pStm, LPDMUS_OBJECTDESC pDesc);
+extern HRESULT IDirectMusicUtils_IPersistStream_ParseUNFOGeneric (DMUS_PRIVATE_CHUNK* pChunk, IStream* pStm, LPDMUS_OBJECTDESC pDesc);
+
 #endif /* __WINE_DMUTILS_H */
