diff --git a/dlls/inetcpl.cpl/inetcpl.rc b/dlls/inetcpl.cpl/inetcpl.rc
--- a/dlls/inetcpl.cpl/inetcpl.rc
+++ b/dlls/inetcpl.cpl/inetcpl.rc
@@ -141,6 +141,10 @@
 
 /* Same Version as IE 8 on WinXP */
 #define WINE_FILEVERSION      8,0,6001,18939
+#define WINE_FILEVERSION_STR "8.00.6001.18939"
+
+#define WINE_PRODUCTVERSION      8,0,6001,18939
+#define WINE_PRODUCTVERSION_STR "8.00.6001.18939"
 
 #include "wine/wine_common_ver.rc"
 
