diff --git a/dlls/mshtml/htmlanchor.c b/dlls/mshtml/htmlanchor.c
--- a/dlls/mshtml/htmlanchor.c
+++ b/dlls/mshtml/htmlanchor.c
@@ -846,8 +846,8 @@
 };
 
 dispex_static_data_t HTMLAnchorElement_dispex = {
-    .id           = OBJID_HTMLAnchorElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLAnchorElement,
+    .prototype_id = PROT_HTMLElement,
     .vtbl         = &HTMLAnchorElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLAnchorElement_tid,
     .iface_tids   = HTMLAnchorElement_iface_tids,
