diff --git a/dlls/mshtml/htmlarea.c b/dlls/mshtml/htmlarea.c
--- a/dlls/mshtml/htmlarea.c
+++ b/dlls/mshtml/htmlarea.c
@@ -456,8 +456,8 @@
     0
 };
 dispex_static_data_t HTMLAreaElement_dispex = {
-    .id           = OBJID_HTMLAreaElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLAreaElement,
+    .prototype_id = PROT_HTMLElement,
     .vtbl         = &HTMLAreaElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLAreaElement_tid,
     .iface_tids   = HTMLAreaElement_iface_tids,
