diff --git a/dlls/mshtml/htmlcurstyle.c b/dlls/mshtml/htmlcurstyle.c
--- a/dlls/mshtml/htmlcurstyle.c
+++ b/dlls/mshtml/htmlcurstyle.c
@@ -1195,8 +1195,8 @@
     0
 };
 dispex_static_data_t MSCurrentStyleCSSProperties_dispex = {
-    .id           = OBJID_MSCurrentStyleCSSProperties,
-    .prototype_id = OBJID_MSCSSProperties,
+    .id           = PROT_MSCurrentStyleCSSProperties,
+    .prototype_id = PROT_MSCSSProperties,
     .vtbl         = &MSCurrentStyleCSSProperties_dispex_vtbl,
     .disp_tid     = DispHTMLCurrentStyle_tid,
     .iface_tids   = MSCurrentStyleCSSProperties_iface_tids,
