diff --git a/dlls/mshtml/htmlgeneric.c b/dlls/mshtml/htmlgeneric.c
--- a/dlls/mshtml/htmlgeneric.c
+++ b/dlls/mshtml/htmlgeneric.c
@@ -114,8 +114,8 @@
 };
 
 dispex_static_data_t HTMLUnknownElement_dispex = {
-    .id           = OBJID_HTMLUnknownElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLUnknownElement,
+    .prototype_id = PROT_HTMLElement,
     .vtbl         = &HTMLGenericElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLGenericElement_tid,
     .iface_tids   = HTMLUnknownElement_iface_tids,
