diff --git a/dlls/mshtml/navigate.c b/dlls/mshtml/navigate.c
--- a/dlls/mshtml/navigate.c
+++ b/dlls/mshtml/navigate.c
@@ -1150,9 +1150,8 @@
                 break;
             case BOM_UTF16:
                 This->nschannel->charset = strdup("utf-16");
-                break;
             case BOM_NONE:
-                break;
+                /* FIXME: Get charset from HTTP headers */;
             }
 
             if(!This->nschannel->content_type) {
