diff --git a/dlls/msxml3/dispex.c b/dlls/msxml3/dispex.c
--- a/dlls/msxml3/dispex.c
+++ b/dlls/msxml3/dispex.c
@@ -102,8 +102,8 @@
     { &IID_IXMLDOMSchemaCollection2, LibXml2 },
     { &IID_IXMLDOMSelection, LibXml2 },
     { &IID_IXMLDOMText, LibXml2 },
-    { &IID_IXMLElement2, LibXml },
-    { &IID_IXMLDocument2, LibXml },
+    { &IID_IXMLElement, LibXml },
+    { &IID_IXMLDocument, LibXml },
     { &IID_IXMLHTTPRequest, LibXml2 },
     { &IID_IXSLProcessor, LibXml2 },
     { &IID_IXSLTemplate, LibXml2 },
