diff --git a/dlls/msxml3/msxml_dispex.h b/dlls/msxml3/msxml_dispex.h
--- a/dlls/msxml3/msxml_dispex.h
+++ b/dlls/msxml3/msxml_dispex.h
@@ -56,8 +56,8 @@
     IXMLDOMSchemaCollection2_tid,
     IXMLDOMSelection_tid,
     IXMLDOMText_tid,
-    IXMLElement2_tid,
-    IXMLDocument2_tid,
+    IXMLElement_tid,
+    IXMLDocument_tid,
     IXMLHTTPRequest_tid,
     IXSLProcessor_tid,
     IXSLTemplate_tid,
