diff --git a/dlls/oleaut32/tests/varformat.c b/dlls/oleaut32/tests/varformat.c
--- a/dlls/oleaut32/tests/varformat.c
+++ b/dlls/oleaut32/tests/varformat.c
@@ -363,10 +363,6 @@
   VARFMT(VT_R8,V_R8,1.0001e-27,"##00.0000e-0",S_OK,"1000.1000e-30");
   VARFMT(VT_R8,V_R8,47.11,".0000E+0",S_OK,".4711E+2");
   VARFMT(VT_R8,V_R8,3.0401e-13,"#####.####e-0%",S_OK,"30401.e-15%");
-  VARFMT(VT_R8,V_R8,29.565331,"0.#####E-#",S_OK,"2.95653E1");
-  VARFMT(VT_R8,V_R8,29.123456789,"0.#####E-###",S_OK,"2.91235E1");
-  VARFMT(VT_R8,V_R8,29.123456789,"0.#####E-#0#0",S_OK,"2.91235E01");
-  VARFMT(VT_R8,V_R8,29.123456789,"0.#####E-0#0#",S_OK,"2.91235E01");
   VARFMT(VT_R8,V_R8,1.57,"0.00",S_OK,"1.57");
   VARFMT(VT_R8,V_R8,-1.57,"0.00",S_OK,"-1.57");
   VARFMT(VT_R8,V_R8,-1.57,"#.##",S_OK,"-1.57");
