diff --git a/dlls/qmgr/tests/qmgr.c b/dlls/qmgr/tests/qmgr.c
--- a/dlls/qmgr/tests/qmgr.c
+++ b/dlls/qmgr/tests/qmgr.c
@@ -109,7 +109,9 @@
 
     wsprintfW(cmdline, L"%s qmgr %s", progname, secret);
     ok(CreateProcessW(NULL, cmdline, NULL, NULL, FALSE, 0L, NULL, NULL, &startup, &info), "CreateProcess\n");
-    wait_child_process(&info);
+    wait_child_process(info.hProcess);
+    ok(CloseHandle(info.hProcess), "CloseHandle\n");
+    ok(CloseHandle(info.hThread), "CloseHandle\n");
 }
 
 static void do_child(const char *secretA)
