diff --git a/dlls/ucrtbase/tests/environ.c b/dlls/ucrtbase/tests/environ.c
--- a/dlls/ucrtbase/tests/environ.c
+++ b/dlls/ucrtbase/tests/environ.c
@@ -312,7 +312,9 @@
     snprintf( tmp, sizeof(tmp), "%s %s create", argv[0], argv[1] );
     ret = CreateProcessA( NULL, tmp, NULL, NULL, FALSE, CREATE_UNICODE_ENVIRONMENT, env, NULL, &si, &pi );
     ok( ret, "Couldn't create child process %s\n", tmp );
-    winetest_wait_child_process( &pi );
+    winetest_wait_child_process( pi.hProcess );
+    CloseHandle( pi.hProcess );
+    CloseHandle( pi.hThread );
     free( env );
 }
 
