diff --git a/dlls/windows.perception.stub/holographic_space.c b/dlls/windows.perception.stub/holographic_space.c
--- a/dlls/windows.perception.stub/holographic_space.c
+++ b/dlls/windows.perception.stub/holographic_space.c
@@ -361,6 +361,10 @@
         *holographic_space = NULL;
         return E_INVALIDARG;
     }
+
+    FIXME( "HACK: Setting WS_EX_NOACTIVATE for %p.\n", window );
+    SetWindowLongW( window, GWL_EXSTYLE, WS_EX_NOACTIVATE );
+
     if (!(impl = calloc( 1, sizeof( *impl ) ))) return E_OUTOFMEMORY;
 
     impl->IHolographicSpace_iface.lpVtbl = &holographic_space_vtbl;
