diff --git a/dlls/windowscodecs/version.rc b/dlls/windowscodecs/version.rc
--- a/dlls/windowscodecs/version.rc
+++ b/dlls/windowscodecs/version.rc
@@ -18,6 +18,9 @@
 
 #define WINE_FILEDESCRIPTION_STR "Wine Windows Codecs Library"
 #define WINE_FILENAME_STR "windowscodecs.dll"
-#define WINE_FILEVERSION 6,0,6001,17009
+#define WINE_FILEVERSION 6,0,6001,170099
+#define WINE_FILEVERSION_STR "6.0.6001.17009"
+#define WINE_PRODUCTVERSION 6,0,6001,170099
+#define WINE_PRODUCTVERSION_STR "6.0.6001.17009"
 
 #include "wine/wine_common_ver.rc"
