diff --git a/dlls/wineps.drv/afm.c b/dlls/wineps.drv/afm.c
--- a/dlls/wineps.drv/afm.c
+++ b/dlls/wineps.drv/afm.c
@@ -170,6 +170,10 @@
                     afmle->afm->FontName, afmle->afm->NumofMetrics,
                     debugstr_w(afmle->afm->EncodingScheme));
 
+	    /* Uncomment to regenerate font data; see afm2c.c */
+
+	    /* PSDRV_AFM2C(afmle->afm); */
+
 #if 0
 	    for (i = 0; i < afmle->afm->NumofMetrics; ++i)
 	    {
