diff --git a/dlls/cryptowinrt/credentials.c b/dlls/cryptowinrt/credentials.c
--- a/dlls/cryptowinrt/credentials.c
+++ b/dlls/cryptowinrt/credentials.c
@@ -118,10 +118,8 @@
 
 DEFINE_IINSPECTABLE( credentials_statics, IKeyCredentialManagerStatics, struct credentials_statics, IActivationFactory_iface );
 
-static HRESULT is_supported_async( IUnknown *invoker, IUnknown *param, PROPVARIANT *result, BOOL called_async )
+static HRESULT WINAPI is_supported_async( IUnknown *invoker, IUnknown *param, PROPVARIANT *result )
 {
-    if (!called_async) return STATUS_PENDING;
-
     result->vt = VT_BOOL;
     result->boolVal = FALSE;
     return S_OK;
