diff --git a/dlls/hvsimanagementapi/classes.idl b/dlls/hvsimanagementapi/classes.idl
--- a/dlls/hvsimanagementapi/classes.idl
+++ b/dlls/hvsimanagementapi/classes.idl
@@ -19,10 +19,5 @@
  */
 
 #pragma makedep register
-#pragma winrt ns_prefix
 
-import "windows.security.isolation.idl";
-
-namespace Windows.Security.Isolation {
-    runtimeclass IsolatedWindowsEnvironmentHost;
-}
+#include "windows.security.isolation.idl"
