diff --git a/dlls/mf/topology_loader.c b/dlls/mf/topology_loader.c
--- a/dlls/mf/topology_loader.c
+++ b/dlls/mf/topology_loader.c
@@ -383,7 +383,10 @@
 
     for (i = 0; SUCCEEDED(hr = IMFMediaTypeHandler_GetMediaTypeByIndex(handler, i, &media_type)); i++)
     {
-        if (SUCCEEDED(hr = IMFMediaTypeHandler_IsMediaTypeSupported(handler, media_type, NULL)))
+        /* HACK: Force initialize media type here, this is now something the topology laoder should do
+         * according to conformance tests but it should hopefully going to solve uninitialized audio
+         * renderer issues. */
+        if (SUCCEEDED(hr = IMFMediaTypeHandler_SetCurrentMediaType(handler, media_type)))
         {
             *type = media_type;
             return hr;
