diff --git a/dlls/sapi/sapi.rgs b/dlls/sapi/sapi.rgs
--- a/dlls/sapi/sapi.rgs
+++ b/dlls/sapi/sapi.rgs
@@ -32,10 +32,7 @@
                 {
                     NoRemove Tokens
                     {
-                        NoRemove 'Wine Default Voice'
-                        {
-                            NoRemove Attributes
-                        }
+                        Delete 'Wine Default Voice'
                     }
                 }
             }
