diff --git a/dlls/windows.networking.connectivity/classes.idl b/dlls/windows.networking.connectivity/classes.idl
--- a/dlls/windows.networking.connectivity/classes.idl
+++ b/dlls/windows.networking.connectivity/classes.idl
@@ -19,10 +19,5 @@
  */
 
 #pragma makedep register
-#pragma winrt ns_prefix
 
-import "windows.networking.connectivity.idl";
-
-namespace Windows.Networking.Connectivity {
-    runtimeclass NetworkInformation;
-}
+#include "windows.networking.connectivity.idl"
