diff --git a/dlls/dmusic/dmobject.h b/dlls/dmusic/dmobject.h
--- a/dlls/dmusic/dmobject.h
+++ b/dlls/dmusic/dmobject.h
@@ -39,12 +39,8 @@
 
 HRESULT stream_chunk_get_array(IStream *stream, const struct chunk_entry *chunk, void **array,
         unsigned int *count, DWORD elem_size);
-HRESULT stream_chunk_get_array_alt(IStream *stream, const struct chunk_entry *chunk, void **array,
-        unsigned int *count, DWORD elem_size, DWORD alt_size);
 HRESULT stream_chunk_get_data(IStream *stream, const struct chunk_entry *chunk, void *data,
         ULONG size);
-HRESULT stream_chunk_get_data_alt(IStream *stream, const struct chunk_entry *chunk, void *data,
-        ULONG size, ULONG alt_size);
 HRESULT stream_chunk_get_wstr(IStream *stream, const struct chunk_entry *chunk, WCHAR *str,
         ULONG size);
 
