diff --git a/dlls/dmusic/wave.c b/dlls/dmusic/wave.c
--- a/dlls/dmusic/wave.c
+++ b/dlls/dmusic/wave.c
@@ -361,10 +361,11 @@
 
     if (!(sample = calloc(1, offsetof(struct sample, loops[1])))) goto failed;
     sample->head.cbSize = sizeof(sample->head);
+    sample->head.cSampleLoops = 1;
     sample->loops[0].ulStart = sf_sample->start_loop - sf_sample->start;
     sample->loops[0].ulLength = sf_sample->end_loop - sf_sample->start_loop;
 
-    data_size = (sf_sample->end - sf_sample->start) * format->nBlockAlign;
+    data_size = sf_sample->end - sf_sample->start;
     if (FAILED(hr = CreateStreamOnHGlobal(NULL, TRUE, &stream))) goto failed;
     offset = sf_sample->start * format->nBlockAlign / format->nChannels;
     if (FAILED(hr = IStream_Write(stream, soundfont->sdta + offset, data_size, &data_size))) goto failed;
