diff --git a/dlls/gdi32/enhmetafile.c b/dlls/gdi32/enhmetafile.c
--- a/dlls/gdi32/enhmetafile.c
+++ b/dlls/gdi32/enhmetafile.c
@@ -3001,12 +3001,3 @@
     DeleteMetaFile(hmf);
     return ret;
 }
-
-/******************************************************************
- *         GetEnhMetaFilePixelFormat   (GDI32.@)
- */
-UINT WINAPI GetEnhMetaFilePixelFormat( HENHMETAFILE hemf, UINT size, PIXELFORMATDESCRIPTOR *fmt )
-{
-    FIXME( "(%p,%u,%p) stub\n", hemf, size, fmt );
-    return GDI_ERROR;
-}
