diff --git a/dlls/gdiplus/tests/brush.c b/dlls/gdiplus/tests/brush.c
--- a/dlls/gdiplus/tests/brush.c
+++ b/dlls/gdiplus/tests/brush.c
@@ -1663,12 +1663,8 @@
         expectf(positions[i], res_positions[i]);
     }
 
-    res_factors[5] = 123.0f;
-    res_positions[5] = 456.0f;
     status = GdipGetPathGradientBlend(brush, res_factors, res_positions, 6);
     expect(Ok, status);
-    ok(res_factors[5] == 123.0f, "Unexpected value %f.\n", res_factors[5]);
-    ok(res_positions[5] == 456.0f, "Unexpected value %f.\n", res_positions[5]);
 
     status = GdipSetPathGradientBlend(brush, factors, positions, 1);
     expect(Ok, status);
