diff --git a/dlls/joy.cpl/Makefile.in b/dlls/joy.cpl/Makefile.in
--- a/dlls/joy.cpl/Makefile.in
+++ b/dlls/joy.cpl/Makefile.in
@@ -1,5 +1,5 @@
 MODULE    = joy.cpl
-IMPORTS   = dxguid dinput dinput8 ole32 comctl32 user32 gdi32 advapi32 xinput combase
+IMPORTS   = dxguid dinput dinput8 ole32 comctl32 user32 gdi32 advapi32 xinput
 
 EXTRADLLFLAGS = -Wb,--prefer-native
 
@@ -8,5 +8,4 @@
 	joy.rc \
 	joy.svg \
 	main.c \
-	wginput.c \
 	xinput.c
