diff --git a/dlls/msxml3/attribute.c b/dlls/msxml3/attribute.c
--- a/dlls/msxml3/attribute.c
+++ b/dlls/msxml3/attribute.c
@@ -377,10 +377,11 @@
     BSTR* p)
 {
     domattr *This = impl_from_IXMLDOMAttribute( iface );
+    static const WCHAR attributeW[] = {'a','t','t','r','i','b','u','t','e',0};
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return return_bstr(L"attribute", p);
+    return return_bstr(attributeW, p);
 }
 
 static HRESULT WINAPI domattr_get_text(
