diff --git a/dlls/shlwapi/tests/string.c b/dlls/shlwapi/tests/string.c
--- a/dlls/shlwapi/tests/string.c
+++ b/dlls/shlwapi/tests/string.c
@@ -1147,12 +1147,6 @@
         memset(wbuf, 0xbf, sizeof(wbuf));
         strret.uType = STRRET_WSTR;
         strret.pOleStr = NULL;
-        hres = pStrRetToBufW(&strret, NULL, wbuf, 10);
-        ok(hres == E_FAIL, "StrRetToBufW returned %08lx\n", hres);
-        ok(!wbuf[0], "StrRetToBuf returned %s\n", wine_dbgstr_w(wbuf));
-
-        memset(wbuf, 0xbf, sizeof(wbuf));
-        strret.uType = 3;
         hres = pStrRetToBufW(&strret, NULL, wbuf, 10);
         ok(hres == E_FAIL, "StrRetToBufW returned %08lx\n", hres);
         ok(!wbuf[0], "StrRetToBuf returned %s\n", wine_dbgstr_w(wbuf));
