diff --git a/dlls/sxs/tests/sxs.c b/dlls/sxs/tests/sxs.c
--- a/dlls/sxs/tests/sxs.c
+++ b/dlls/sxs/tests/sxs.c
@@ -231,7 +231,11 @@
     si.cb = sizeof(si);
     ret = CreateProcessA(exe, cmdline, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
     ok(ret, "Could not create process: %lu\n", GetLastError());
-    wait_child_process(&pi);
+
+    wait_child_process(pi.hProcess);
+
+    CloseHandle(pi.hThread);
+    CloseHandle(pi.hProcess);
 }
 
 static void test_SxsLookupClrGuid(void)
