diff --git a/dlls/wintab32/wintab_internal.h b/dlls/wintab32/wintab_internal.h
--- a/dlls/wintab32/wintab_internal.h
+++ b/dlls/wintab32/wintab_internal.h
@@ -150,6 +150,12 @@
                              LPARAM lParam, BOOL send_always);
 LPOPENCONTEXT AddPacketToContextQueue(LPWTPACKET packet, HWND hwnd);
 
+/* X11drv functions */
+extern int  (CDECL *pLoadTabletInfo)(HWND hwnddefault);
+extern int  (CDECL *pGetCurrentPacket)(LPWTPACKET packet);
+extern int  (CDECL *pAttachEventQueueToTablet)(HWND hOwner);
+extern UINT (CDECL *pWTInfoW)(UINT wCategory, UINT nIndex, LPVOID lpOutput);
+
 extern HWND hwndDefault;
 extern CRITICAL_SECTION csTablet;
 
