diff --git a/dlls/d3dx9_36/tests/core.c b/dlls/d3dx9_36/tests/core.c
--- a/dlls/d3dx9_36/tests/core.c
+++ b/dlls/d3dx9_36/tests/core.c
@@ -1271,13 +1271,6 @@
 
     /* viewport */
     hr = ID3DXRenderToSurface_BeginScene(render, surface, &viewport);
-    if (FAILED(hr))
-    {
-        skip("Failed to create ID3DXRenderToSurface\n");
-        check_release((IUnknown *)surface, 0);
-        ID3DXRenderToSurface_Release(render);
-        return;
-    }
     ok(hr == D3D_OK, "ID3DXRenderToSurface::BeginScene returned %#lx, expected %#lx\n", hr, D3D_OK);
     check_ref((IUnknown *)surface, 2);
     if (SUCCEEDED(hr)) ID3DXRenderToSurface_EndScene(render, D3DX_FILTER_NONE);
