diff --git a/dlls/dssenh/tests/dssenh.c b/dlls/dssenh/tests/dssenh.c
--- a/dlls/dssenh/tests/dssenh.c
+++ b/dlls/dssenh/tests/dssenh.c
@@ -144,13 +144,6 @@
     ok(result, "Expected release of the provider.\n");
 
     result = CryptAcquireContextA(&hProv, NULL, MS_ENH_DSS_DH_PROV_A, PROV_DSS_DH, 0);
-    ok(result, "Expected no errors.\n");
-
-    result = CryptReleaseContext(hProv, 0);
-    ok(result, "Expected release of the provider.\n");
-
-    result = CryptAcquireContextA(&hProv, NULL, MS_ENH_DSS_DH_PROV_A, PROV_DSS_DH,
-        CRYPT_VERIFYCONTEXT | CRYPT_NEWKEYSET);
     ok(result, "Expected no errors.\n");
 
     result = CryptReleaseContext(hProv, 0);
