diff --git a/dlls/mmdevapi/audiosessionmanager.c b/dlls/mmdevapi/audiosessionmanager.c
--- a/dlls/mmdevapi/audiosessionmanager.c
+++ b/dlls/mmdevapi/audiosessionmanager.c
@@ -175,13 +175,6 @@
     return S_OK;
 }
 
-struct session_mgr
-{
-    IAudioSessionManager2 IAudioSessionManager2_iface;
-    IMMDevice *device;
-    LONG ref;
-};
-
 static inline struct session_mgr *impl_from_IAudioSessionManager2(IAudioSessionManager2 *iface)
 {
     return CONTAINING_RECORD(iface, struct session_mgr, IAudioSessionManager2_iface);
