diff --git a/dlls/mshtml/nsevents.c b/dlls/mshtml/nsevents.c
--- a/dlls/mshtml/nsevents.c
+++ b/dlls/mshtml/nsevents.c
@@ -573,13 +573,6 @@
     nsIDOMEventTarget *target;
     unsigned i;
 
-    if(doc->nsevent_listener) {
-        listener = doc->nsevent_listener;
-        listener->doc = NULL;
-        doc->nsevent_listener = NULL;
-        release_listener(listener);
-    }
-
     listener = malloc(sizeof(nsDocumentEventListener));
     if(!listener)
         return;
