diff --git a/dlls/oleaut32/varformat.c b/dlls/oleaut32/varformat.c
--- a/dlls/oleaut32/varformat.c
+++ b/dlls/oleaut32/varformat.c
@@ -690,10 +690,9 @@
       }
       pFormat += 2;
       *++pOut = 0x0;
-      while (*pFormat == '0' || *pFormat == '#')
-      {
-        if (*pFormat == '0')
-          *pOut = *pOut + 1;
+      while (*pFormat == '0')
+      {
+        *pOut = *pOut + 1;
         pFormat++;
       }
       pOut++;
