diff --git a/dlls/webservices/rsrc.rc b/dlls/webservices/rsrc.rc
--- a/dlls/webservices/rsrc.rc
+++ b/dlls/webservices/rsrc.rc
@@ -18,6 +18,12 @@
 
 #define WINE_FILEDESCRIPTION_STR "Wine Web Services"
 #define WINE_FILENAME_STR "webservices.dll"
-#define WINE_FILEVERSION 10,0,14393,0
+#define WINE_FILEVERSION_MAJOR 10
+#define WINE_FILEVERSION_MINOR 0
+#define WINE_FILEVERSION_BUILD 14393
+#define WINE_FILEVERSION_PLATFORMID 0
+
+#define WINE_PRODUCTVERSION 10,0,14993,0
+#define WINE_PRODUCTVERSION_STR "10.0.14993.0"
 
 #include "wine/wine_common_ver.rc"
