diff --git a/dlls/compstui/compstui.rc b/dlls/compstui/compstui.rc
--- a/dlls/compstui/compstui.rc
+++ b/dlls/compstui/compstui.rc
@@ -28,14 +28,4 @@
     IDS_CPSUI_PROPERTIES    "Properties"
     IDS_CPSUI_OPTIONS       "Options"
     IDS_CPSUI_DEFAULT       "Default"
-    IDS_CPSUI_COLOR         "C&olor"
-    IDS_CPSUI_COLOR_APPERANCE "Color"
-    IDS_CPSUI_MONOCHROME    "Bla&ck && White"
-    IDS_CPSUI_NUM_OF_COPIES "&Copy count"
-    IDS_CPSUI_DUPLEX        "Print on Both Sides"
-    IDS_CPSUI_SIMPLEX       "&None"
-    IDS_CPSUI_HORIZONTAL    "Flip on Shor&t Edge"
-    IDS_CPSUI_VERTICAL      "Flip on Long E&dge"
-    IDS_CPSUI_LONG_SIDE     "Flip on Long E&dge"
-    IDS_CPSUI_SHORT_SIDE    "Flip on Shor&t Edge"
 }
