diff --git a/dlls/crypt32/oid.c b/dlls/crypt32/oid.c
--- a/dlls/crypt32/oid.c
+++ b/dlls/crypt32/oid.c
@@ -1716,13 +1716,7 @@
 {
     PCCRYPT_OID_INFO ret = NULL;
 
-    TRACE("(%#lx, %p, %lu)\n", dwKeyType, pvKey, dwGroupId);
-
-    if (dwKeyType & (CRYPT_OID_INFO_PUBKEY_ENCRYPT_KEY_FLAG | CRYPT_OID_INFO_PUBKEY_SIGN_KEY_FLAG))
-    {
-        FIXME("flags %#lx not supported\n", dwKeyType & (CRYPT_OID_INFO_PUBKEY_ENCRYPT_KEY_FLAG | CRYPT_OID_INFO_PUBKEY_SIGN_KEY_FLAG));
-        dwKeyType &= ~(CRYPT_OID_INFO_PUBKEY_ENCRYPT_KEY_FLAG | CRYPT_OID_INFO_PUBKEY_SIGN_KEY_FLAG);
-    }
+    TRACE("(%ld, %p, %ld)\n", dwKeyType, pvKey, dwGroupId);
 
     switch(dwKeyType)
     {
