diff --git a/dlls/ieframe/webbrowser.c b/dlls/ieframe/webbrowser.c
--- a/dlls/ieframe/webbrowser.c
+++ b/dlls/ieframe/webbrowser.c
@@ -612,11 +612,13 @@
 
 static HRESULT WINAPI WebBrowser_get_Name(IWebBrowser2 *iface, BSTR *Name)
 {
+    static const WCHAR sName[] = {'M','i','c','r','o','s','o','f','t',' ','W','e','b',
+                                  ' ','B','r','o','w','s','e','r',' ','C','o','n','t','r','o','l',0};
     WebBrowser *This = impl_from_IWebBrowser2(iface);
 
     TRACE("(%p)->(%p)\n", This, Name);
 
-    *Name = SysAllocString(L"Microsoft Web Browser Control");
+    *Name = SysAllocString(sName);
 
     return S_OK;
 }
