diff --git a/dlls/mshtml/htmllocation.c b/dlls/mshtml/htmllocation.c
--- a/dlls/mshtml/htmllocation.c
+++ b/dlls/mshtml/htmllocation.c
@@ -605,10 +605,10 @@
     0
 };
 static dispex_static_data_t HTMLLocation_dispex = {
-    .name             = "Location",
-    .vtbl             = &HTMLLocation_dispex_vtbl,
-    .disp_tid         = DispHTMLLocation_tid,
-    .iface_tids       = HTMLLocation_iface_tids
+    "Location",
+    &HTMLLocation_dispex_vtbl,
+    DispHTMLLocation_tid,
+    HTMLLocation_iface_tids
 };
 
 HRESULT create_location(HTMLOuterWindow *window, HTMLLocation **ret)
