diff --git a/dlls/mssign32/mssign32_main.c b/dlls/mssign32/mssign32_main.c
--- a/dlls/mssign32/mssign32_main.c
+++ b/dlls/mssign32/mssign32_main.c
@@ -25,6 +25,7 @@
 #include "wincrypt.h"
 
 #include "wine/debug.h"
+#include "wine/heap.h"
 #include "wine/mssign.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(mssign);
@@ -77,5 +78,6 @@
 
 HRESULT WINAPI SignerFreeSignerContext(SIGNER_CONTEXT *signer_context)
 {
+    heap_free(signer_context);
     return S_OK;
 }
