diff --git a/dlls/ntoskrnl.exe/ntoskrnl_private.h b/dlls/ntoskrnl.exe/ntoskrnl_private.h
--- a/dlls/ntoskrnl.exe/ntoskrnl_private.h
+++ b/dlls/ntoskrnl.exe/ntoskrnl_private.h
@@ -22,7 +22,6 @@
 #define __WINE_NTOSKRNL_PRIVATE_H
 
 #include <stdarg.h>
-#include <stdbool.h>
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
 #include "windef.h"
@@ -121,6 +120,5 @@
 {
     DEVICE_OBJECT device_obj;
     DEVICE_RELATIONS *children;
-    HKEY dyn_data_key;
 };
 #endif
