diff --git a/dlls/spoolss/spoolss_main.c b/dlls/spoolss/spoolss_main.c
--- a/dlls/spoolss/spoolss_main.c
+++ b/dlls/spoolss/spoolss_main.c
@@ -91,12 +91,12 @@
 /******************************************************************
  *   BuildOtherNamesFromMachineName   [SPOOLSS.@]
  */
-BOOL WINAPI BuildOtherNamesFromMachineName(WCHAR ***ptr1, DWORD *ptr2)
+BOOL WINAPI BuildOtherNamesFromMachineName(LPVOID * ptr1, LPVOID * ptr2)
 {
     FIXME("(%p, %p) stub\n", ptr1, ptr2);
 
     *ptr1 = NULL;
-    *ptr2 = 0;
+    *ptr2 = NULL;
     return FALSE;
 }
 
