diff --git a/dlls/windowscodecs/tests/pngformat.c b/dlls/windowscodecs/tests/pngformat.c
--- a/dlls/windowscodecs/tests/pngformat.c
+++ b/dlls/windowscodecs/tests/pngformat.c
@@ -314,12 +314,13 @@
 
 static WCHAR *save_profile( BYTE *buffer, UINT size )
 {
+    static const WCHAR tstW[] = {'t','s','t',0};
     WCHAR path[MAX_PATH], filename[MAX_PATH], *ret;
     HANDLE handle;
     DWORD count;
 
     GetTempPathW(MAX_PATH, path);
-    GetTempFileNameW(path, L"tst", 0, filename);
+    GetTempFileNameW(path, tstW, 0, filename);
 
     handle = CreateFileW(filename, GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, 0, NULL);
     if (handle == INVALID_HANDLE_VALUE) return NULL;
