diff --git a/dlls/dmusic/band.c b/dlls/dmusic/band.c
--- a/dlls/dmusic/band.c
+++ b/dlls/dmusic/band.c
@@ -239,9 +239,12 @@
         switch (MAKE_IDTYPE(chunk.id, chunk.type))
         {
         case DMUS_FOURCC_INSTRUMENT_CHUNK:
-            hr = stream_chunk_get_data_alt(stream, &chunk, &inst, sizeof(inst),
-                    offsetof(DMUS_IO_INSTRUMENT, nPitchBendRange));
-            break;
+        {
+            UINT size = sizeof(inst);
+            if (chunk.size == offsetof(DMUS_IO_INSTRUMENT, nPitchBendRange)) size = chunk.size;
+            if (FAILED(hr = stream_chunk_get_data(stream, &chunk, &inst, size))) break;
+            break;
+        }
 
         case MAKE_IDTYPE(FOURCC_LIST, DMUS_FOURCC_REF_LIST):
         {
