diff --git a/dlls/jscript/error.c b/dlls/jscript/error.c
--- a/dlls/jscript/error.c
+++ b/dlls/jscript/error.c
@@ -479,7 +479,6 @@
         case JS_E_MISSING_ARG:
         case JS_E_OBJECT_NOT_COLLECTION:
         case JS_E_FUNCTION_EXPECTED:
-        case JS_E_STRING_EXPECTED:
         case JS_E_DATE_EXPECTED:
         case JS_E_NUMBER_EXPECTED:
         case JS_E_OBJECT_EXPECTED:
@@ -490,7 +489,6 @@
         case JS_E_JSCRIPT_EXPECTED:
         case JS_E_ENUMERATOR_EXPECTED:
         case JS_E_REGEXP_EXPECTED:
-        case JS_E_ARRAY_OR_ARGS_EXPECTED:
         case JS_E_ARRAY_EXPECTED:
         case JS_E_CYCLIC_PROTO_VALUE:
         case JS_E_CANNOT_CREATE_FOR_NONEXTENSIBLE:
