diff --git a/dlls/mmdevapi/unixlib.h b/dlls/mmdevapi/unixlib.h
--- a/dlls/mmdevapi/unixlib.h
+++ b/dlls/mmdevapi/unixlib.h
@@ -136,6 +136,7 @@
     EDataFlow flow;
     AUDCLNT_SHAREMODE share;
     const WAVEFORMATEX *fmt_in;
+    WAVEFORMATEXTENSIBLE *fmt_out;
     HRESULT result;
 };
 
@@ -343,7 +344,6 @@
     test_connect,
     is_started,
     get_prop_value,
-    midi_get_driver,
     midi_init,
     midi_release,
     midi_out_message,
