diff --git a/dlls/ntdll/wcstring.c b/dlls/ntdll/wcstring.c
--- a/dlls/ntdll/wcstring.c
+++ b/dlls/ntdll/wcstring.c
@@ -352,11 +352,10 @@
 #undef wcsncpy
 LPWSTR __cdecl wcsncpy( LPWSTR s1, LPCWSTR s2, size_t n )
 {
-    size_t i;
-
-    for (i = 0; i < n; i++) if(!(s1[i] = s2[i])) break;
-    for (; i < n; i++) s1[i] = 0;
-    return s1;
+    WCHAR *ret = s1;
+    for ( ; n; n--) if (!(*s1++ = *s2++)) break;
+    for ( ; n; n--) *s1++ = 0;
+    return ret;
 }
 
 
