diff --git a/dlls/shell32/Makefile.in b/dlls/shell32/Makefile.in
--- a/dlls/shell32/Makefile.in
+++ b/dlls/shell32/Makefile.in
@@ -21,7 +21,6 @@
 	dragdrophelper.c \
 	ebrowser.c \
 	enumidlist.c \
-	enumobjects.c \
 	folders.c \
 	iconcache.c \
 	new_menu.c \
@@ -61,7 +60,6 @@
 	resources/restricted.svg \
 	resources/search.svg \
 	resources/security.svg \
-	resources/shield.svg \
 	resources/shortcut.svg \
 	resources/shut_down.svg \
 	resources/sleep.svg \
