diff --git a/dlls/win32u/dibdrv/dibdrv.h b/dlls/win32u/dibdrv/dibdrv.h
--- a/dlls/win32u/dibdrv/dibdrv.h
+++ b/dlls/win32u/dibdrv/dibdrv.h
@@ -289,3 +289,15 @@
 {
     return dib->color_table ? dib->color_table : get_default_color_table( dib->bit_count );
 }
+
+struct osmesa_funcs
+{
+    void (*get_gl_funcs)( struct opengl_funcs *funcs );
+    struct wgl_context * (*create_context)( HDC hdc, const PIXELFORMATDESCRIPTOR *descr );
+    BOOL (*delete_context)( struct wgl_context *context );
+    PROC (*get_proc_address)( const char *proc );
+    BOOL (*make_current)( struct wgl_context *context, void *bits,
+                          int width, int height, int bpp, int stride );
+};
+
+extern const struct osmesa_funcs *init_opengl_lib(void);
