diff --git a/dlls/windows.gaming.input/private.h b/dlls/windows.gaming.input/private.h
--- a/dlls/windows.gaming.input/private.h
+++ b/dlls/windows.gaming.input/private.h
@@ -43,7 +43,6 @@
 #include "wine/list.h"
 
 #include "provider.h"
-#include "async_private.h"
 
 extern HINSTANCE windows_gaming_input;
 extern ICustomGameControllerFactory *controller_factory;
@@ -77,6 +76,7 @@
 extern HRESULT force_feedback_motor_create( IDirectInputDevice8W *device, IForceFeedbackMotor **out );
 extern HRESULT force_feedback_effect_create( enum WineForceFeedbackEffectType type, IInspectable *outer, IWineForceFeedbackEffectImpl **out );
 
+typedef HRESULT (WINAPI *async_operation_callback)( IUnknown *invoker, IUnknown *param, PROPVARIANT *result );
 extern HRESULT async_operation_boolean_create( IUnknown *invoker, IUnknown *param, async_operation_callback callback,
                                                IAsyncOperation_boolean **out );
 extern HRESULT async_operation_effect_result_create( IUnknown *invoker, IUnknown *param, async_operation_callback callback,
