diff --git a/dlls/windows.media/classes.idl b/dlls/windows.media/classes.idl
--- a/dlls/windows.media/classes.idl
+++ b/dlls/windows.media/classes.idl
@@ -19,14 +19,5 @@
  */
 
 #pragma makedep register
-#pragma winrt ns_prefix
 
-import "windows.media.closedcaptioning.idl";
-import "windows.media.transcoding.idl";
-
-namespace Windows.Media.ClosedCaptioning {
-    runtimeclass ClosedCaptionProperties;
-}
-namespace Windows.Media.Transcoding {
-    runtimeclass MediaTranscoder;
-}
+#include "windows.media.closedcaptioning.idl"
