diff --git a/dlls/dplayx/dplayx_messages.h b/dlls/dplayx/dplayx_messages.h
--- a/dlls/dplayx/dplayx_messages.h
+++ b/dlls/dplayx/dplayx_messages.h
@@ -74,7 +74,7 @@
 #define DPMSG_DEFAULT_WAIT_TIME DPMSG_WAIT_30_SECS
 
 /* Message types etc. */
-#pragma pack(push,1)
+#include "pshpack1.h"
 
 typedef struct
 {
@@ -333,6 +333,6 @@
   DPID id;
 } DPSP_MSG_ADDFORWARDACK;
 
-#pragma pack(pop)
+#include "poppack.h"
 
 #endif
