diff --git a/dlls/hhctrl.ocx/search.c b/dlls/hhctrl.ocx/search.c
--- a/dlls/hhctrl.ocx/search.c
+++ b/dlls/hhctrl.ocx/search.c
@@ -154,6 +154,7 @@
 static SearchItem *SearchCHM_Storage(SearchItem *item, IStorage *pStorage,
                                      const char *needle)
 {
+    static const WCHAR szHTMext[] = {'.','h','t','m',0};
     IEnumSTATSTG *elem = NULL;
     WCHAR *filename = NULL;
     STATSTG entries;
@@ -176,7 +177,7 @@
             item = SearchCHM_Folder(item, pStorage, filename, needle);
             break;
         case STGTY_STREAM:
-            if(wcsstr(filename, L".htm"))
+            if(wcsstr(filename, szHTMext))
             {
                 WCHAR *title = SearchCHM_File(pStorage, filename, needle);
 
