diff --git a/dlls/msvcrt/tests/misc.c b/dlls/msvcrt/tests/misc.c
--- a/dlls/msvcrt/tests/misc.c
+++ b/dlls/msvcrt/tests/misc.c
@@ -25,7 +25,6 @@
 #include <stdio.h>
 #include <math.h>
 #include <process.h>
-#include <sys/stat.h>
 
 static inline BOOL almost_equal(double d1, double d2) {
     if(d1-d2>-1e-30 && d1-d2<1e-30)
@@ -365,7 +364,7 @@
 
     tempf = _tempnam(".", "wne");
     ok(tempf != NULL, "_tempnam failed\n");
-    fd = _open(tempf, _O_CREAT | _O_WRONLY, _S_IWRITE);
+    fd = _open(tempf, _O_CREAT | _O_WRONLY);
     ok(fd != -1, "open failed\n");
 
     sprintf(buf, "\"%s\" misc popen %d", name, fd);
