diff --git a/dlls/shell32/tests/brsfolder.c b/dlls/shell32/tests/brsfolder.c
--- a/dlls/shell32/tests/brsfolder.c
+++ b/dlls/shell32/tests/brsfolder.c
@@ -32,7 +32,6 @@
 #define TIMER_WAIT_MS 50 /* Should be long enough for slow systems */
 
 static const char new_folder_name[] = "foo";
-static const WCHAR new_folder_name_w[] = L"foo";
 static LPITEMIDLIST selected_folder_pidl;
 
 /*
@@ -295,7 +294,7 @@
 
         ret = SendMessageA(hwnd, BFFM_SETSELECTIONA, 1, (LPARAM)new_folder_name);
         ok(!ret, "SendMessage returned: %lu\n", ret);
-        ret = SendMessageW(hwnd, BFFM_SETSELECTIONW, 1, (LPARAM)new_folder_name_w);
+        ret = SendMessageW(hwnd, BFFM_SETSELECTIONW, 1, (LPARAM)new_folder_name);
         ok(!ret, "SendMessage returned: %lu\n", ret);
 
         SendMessageA(hwnd, WM_COMMAND, IDOK, 0);
