diff --git a/dlls/shell32/tests/shlexec.c b/dlls/shell32/tests/shlexec.c
--- a/dlls/shell32/tests/shlexec.c
+++ b/dlls/shell32/tests/shlexec.c
@@ -1388,6 +1388,7 @@
 
 static void test_commandline2argv(void)
 {
+    static const WCHAR exeW[] = {'e','x','e',0};
     const cmdline_tests_t* test;
     WCHAR strW[MAX_PATH];
     LPWSTR *args;
@@ -1403,7 +1404,7 @@
     }
 
     SetLastError(0xdeadbeef);
-    args = CommandLineToArgvW(L"exe", NULL);
+    args = CommandLineToArgvW(exeW, NULL);
     le = GetLastError();
     ok(args == NULL && le == ERROR_INVALID_PARAMETER, "expected NULL with ERROR_INVALID_PARAMETER got %p with %lu\n", args, le);
 
