diff --git a/dlls/urlmon/urlmon_main.c b/dlls/urlmon/urlmon_main.c
--- a/dlls/urlmon/urlmon_main.c
+++ b/dlls/urlmon/urlmon_main.c
@@ -351,8 +351,6 @@
     CF_LockServer
 };
 
-static ClassFactory PersistentZoneIdentifierCF =
-    { { &ClassFactoryVtbl }, PersistentZoneIdentifier_Construct};
 static ClassFactory FileProtocolCF =
     { { &ClassFactoryVtbl }, FileProtocol_Construct};
 static ClassFactory FtpProtocolCF =
@@ -385,7 +383,6 @@
 
 static const struct object_creation_info object_creation[] =
 {
-    { &CLSID_PersistentZoneIdentifier,            &PersistentZoneIdentifierCF.IClassFactory_iface,    NULL },
     { &CLSID_FileProtocol,            &FileProtocolCF.IClassFactory_iface,    L"file" },
     { &CLSID_FtpProtocol,             &FtpProtocolCF.IClassFactory_iface,     L"ftp"  },
     { &CLSID_GopherProtocol,          &GopherProtocolCF.IClassFactory_iface,  L"gopher" },
