diff --git a/dlls/commdlg.dll16/filedlg.c b/dlls/commdlg.dll16/filedlg.c
--- a/dlls/commdlg.dll16/filedlg.c
+++ b/dlls/commdlg.dll16/filedlg.c
@@ -496,7 +496,7 @@
 }
 
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 struct hook_proc
 {
     BYTE popl_eax;    /* popl %eax */
@@ -506,7 +506,7 @@
     BYTE jmp;         /* jmp call_hook */
     DWORD call_hook;
 };
-#pragma pack(pop)
+#include "poppack.h"
 
 static LPOFNHOOKPROC alloc_hook( LPOFNHOOKPROC16 hook16 )
 {
