diff --git a/dlls/d3d11/utils.c b/dlls/d3d11/utils.c
--- a/dlls/d3d11/utils.c
+++ b/dlls/d3d11/utils.c
@@ -311,7 +311,7 @@
         case WINED3DFMT_BC7_TYPELESS: return DXGI_FORMAT_BC7_TYPELESS;
         case WINED3DFMT_BC7_UNORM: return DXGI_FORMAT_BC7_UNORM;
         case WINED3DFMT_BC7_UNORM_SRGB: return DXGI_FORMAT_BC7_UNORM_SRGB;
-        case WINED3DFMT_NV12_PLANAR: return DXGI_FORMAT_NV12;
+        case WINED3DFMT_NV12: return DXGI_FORMAT_NV12;
         case WINED3DFMT_YUY2: return DXGI_FORMAT_YUY2;
         case WINED3DFMT_B4G4R4A4_UNORM: return DXGI_FORMAT_B4G4R4A4_UNORM;
         default:
@@ -424,7 +424,7 @@
         case DXGI_FORMAT_BC7_TYPELESS: return WINED3DFMT_BC7_TYPELESS;
         case DXGI_FORMAT_BC7_UNORM: return WINED3DFMT_BC7_UNORM;
         case DXGI_FORMAT_BC7_UNORM_SRGB: return WINED3DFMT_BC7_UNORM_SRGB;
-        case DXGI_FORMAT_NV12: return WINED3DFMT_NV12_PLANAR;
+        case DXGI_FORMAT_NV12: return WINED3DFMT_NV12;
         case DXGI_FORMAT_YUY2: return WINED3DFMT_YUY2;
         case DXGI_FORMAT_B4G4R4A4_UNORM: return WINED3DFMT_B4G4R4A4_UNORM;
         default:
