diff --git a/dlls/gdiplus/font.c b/dlls/gdiplus/font.c
--- a/dlls/gdiplus/font.c
+++ b/dlls/gdiplus/font.c
@@ -34,7 +34,7 @@
 #include "gdiplus_private.h"
 
 /* PANOSE is 10 bytes in size, need to pack the structure properly */
-#pragma pack(push,2)
+#include "pshpack2.h"
 typedef struct
 {
     USHORT version;
@@ -99,7 +99,7 @@
     SHORT metricDataFormat;
     USHORT numberOfHMetrics;
 } TT_HHEA;
-#pragma pack(pop)
+#include "poppack.h"
 
 #ifdef WORDS_BIGENDIAN
 #define GET_BE_WORD(x) (x)
