diff --git a/dlls/hidclass.sys/hid.h b/dlls/hidclass.sys/hid.h
--- a/dlls/hidclass.sys/hid.h
+++ b/dlls/hidclass.sys/hid.h
@@ -47,6 +47,8 @@
     WCHAR instance_id[MAX_DEVICE_ID_LEN];
     WCHAR container_id[MAX_GUID_STRING_LEN];
     const GUID *class_guid;
+
+    HANDLE steam_overlay_event;
 
     BOOL is_fdo;
 };
@@ -133,6 +135,8 @@
 
     PDRIVER_ADD_DEVICE AddDevice;
     PDRIVER_DISPATCH PNPDispatch;
+
+    HANDLE steam_overlay_event;
 } minidriver;
 
 void call_minidriver( ULONG code, DEVICE_OBJECT *device, void *in_buff, ULONG in_size,
