diff --git a/dlls/keyboard.drv16/keyboard.c b/dlls/keyboard.drv16/keyboard.c
--- a/dlls/keyboard.drv16/keyboard.c
+++ b/dlls/keyboard.drv16/keyboard.c
@@ -37,7 +37,7 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(keyboard);
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 typedef struct _KBINFO
 {
     BYTE Begin_First_Range;
@@ -46,7 +46,7 @@
     BYTE End_Second_Range;
     WORD StateSize;
 } KBINFO, *LPKBINFO;
-#pragma pack(pop)
+#include "poppack.h"
 
 static FARPROC16 DefKeybEventProc;
 static LPBYTE pKeyStateTable;
