diff --git a/dlls/mouse.drv16/mouse.c b/dlls/mouse.drv16/mouse.c
--- a/dlls/mouse.drv16/mouse.c
+++ b/dlls/mouse.drv16/mouse.c
@@ -27,7 +27,7 @@
 #include "wine/winbase16.h"
 #include "wine/winuser16.h"
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 typedef struct _MOUSEINFO
 {
     BYTE msExist;
@@ -40,7 +40,7 @@
     WORD msYRes;
     WORD msMouseCommPort;
 } MOUSEINFO, *LPMOUSEINFO;
-#pragma pack(pop)
+#include "poppack.h"
 
 static FARPROC16 DefMouseEventProc;
 
