diff --git a/dlls/mshtml/range.c b/dlls/mshtml/range.c
--- a/dlls/mshtml/range.c
+++ b/dlls/mshtml/range.c
@@ -1685,7 +1685,7 @@
     0
 };
 dispex_static_data_t TextRange_dispex = {
-    .id         = OBJID_TextRange,
+    .id         = PROT_TextRange,
     .vtbl       = &HTMLTxtRange_dispex_vtbl,
     .disp_tid   = IHTMLTxtRange_tid,
     .iface_tids = TextRange_iface_tids,
@@ -1991,7 +1991,7 @@
 };
 
 dispex_static_data_t Range_dispex = {
-    .id         = OBJID_Range,
+    .id         = PROT_Range,
     .vtbl       = &HTMLDOMRange_dispex_vtbl,
     .disp_tid   = DispHTMLDOMRange_tid,
     .iface_tids = Range_iface_tids,
