diff --git a/dlls/ntdll/tests/time.c b/dlls/ntdll/tests/time.c
--- a/dlls/ntdll/tests/time.c
+++ b/dlls/ntdll/tests/time.c
@@ -156,7 +156,7 @@
 static void test_RtlQueryPerformanceCounter(void)
 {
     struct hypervisor_shared_data *hsd;
-    KUSER_SHARED_DATA *usd = (void *)0x7ffe0000;
+    KSHARED_USER_DATA *usd = (void *)0x7ffe0000;
     LARGE_INTEGER frequency, counter;
     NTSTATUS status;
     UINT64 tsc0, tsc1;
@@ -418,7 +418,7 @@
 
 static void test_user_shared_data_time(void)
 {
-    KUSER_SHARED_DATA *user_shared_data = (void *)0x7ffe0000;
+    KSHARED_USER_DATA *user_shared_data = (void *)0x7ffe0000;
     SYSTEM_TIMEOFDAY_INFORMATION timeofday;
     ULONGLONG t1, t2, t3;
     NTSTATUS status;
