diff --git a/dlls/quartz/memallocator.c b/dlls/quartz/memallocator.c
--- a/dlls/quartz/memallocator.c
+++ b/dlls/quartz/memallocator.c
@@ -162,12 +162,17 @@
 static HRESULT WINAPI BaseMemAllocator_SetProperties(IMemAllocator * iface, ALLOCATOR_PROPERTIES *pRequest, ALLOCATOR_PROPERTIES *pActual)
 {
     BaseMemAllocator *This = impl_from_IMemAllocator(iface);
+    const char *sgi;
     HRESULT hr;
 
     TRACE("(%p)->(%p, %p)\n", This, pRequest, pActual);
 
     TRACE("Requested %ld buffers, size %ld, alignment %ld, prefix %ld.\n",
             pRequest->cBuffers, pRequest->cbBuffer, pRequest->cbAlign, pRequest->cbPrefix);
+
+    sgi = getenv("SteamGameId");
+    if (sgi && (!strcmp(sgi, "1113000")))
+        pRequest->cBuffers = min(8, pRequest->cBuffers);
 
     EnterCriticalSection(This->pCritSect);
     {
