diff --git a/dlls/riched20/tests/txtsrv.c b/dlls/riched20/tests/txtsrv.c
--- a/dlls/riched20/tests/txtsrv.c
+++ b/dlls/riched20/tests/txtsrv.c
@@ -529,7 +529,7 @@
 
 static void *wrapperCodeMem = NULL;
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 
 /* Code structure for x86 byte code */
 typedef struct
@@ -551,7 +551,7 @@
     int  vtablefunc_offset;
 } STDCALL_TO_THISCALL_THUNK;
 
-#pragma pack(pop)
+#include "poppack.h"
 
 static void setup_thiscall_wrappers(void)
 {
