diff --git a/dlls/rsaenh/Makefile.in b/dlls/rsaenh/Makefile.in
--- a/dlls/rsaenh/Makefile.in
+++ b/dlls/rsaenh/Makefile.in
@@ -1,10 +1,15 @@
 MODULE    = rsaenh.dll
 IMPORTLIB = rsaenh
-IMPORTS   = $(TOMCRYPT_PE_LIBS) crypt32 advapi32
-EXTRAINCL = $(TOMCRYPT_PE_CFLAGS)
+IMPORTS   = bcrypt crypt32 advapi32
 
 SOURCES = \
+	aes.c \
+	des.c \
 	handle.c \
 	implglue.c \
+	mpi.c \
+	rc2.c \
+	rc4.c \
+	rsa.c \
 	rsaenh.c \
 	rsrc.rc
