diff --git a/dlls/schedsvc/atsvc.c b/dlls/schedsvc/atsvc.c
--- a/dlls/schedsvc/atsvc.c
+++ b/dlls/schedsvc/atsvc.c
@@ -860,7 +860,7 @@
 {
     HANDLE hfile;
     DWORD try, size;
-#pragma pack(push,2)
+#include "pshpack2.h"
     struct
     {
         UINT exit_code;
@@ -869,7 +869,7 @@
         SYSTEMTIME last_runtime;
         WORD instance_count;
     } state;
-#pragma pack(pop)
+#include "poppack.h"
 
     try = 1;
     for (;;)
