diff --git a/dlls/secur32/schannel_gnutls.c b/dlls/secur32/schannel_gnutls.c
--- a/dlls/secur32/schannel_gnutls.c
+++ b/dlls/secur32/schannel_gnutls.c
@@ -1600,14 +1600,9 @@
 
 static NTSTATUS process_detach( void *args )
 {
-    if (libgnutls_handle)
-    {
-        if (TRACE_ON(secur32))
-            pgnutls_global_set_log_function(NULL);
-        pgnutls_global_deinit();
-        dlclose(libgnutls_handle);
-        libgnutls_handle = NULL;
-    }
+    pgnutls_global_deinit();
+    dlclose(libgnutls_handle);
+    libgnutls_handle = NULL;
     return STATUS_SUCCESS;
 }
 
