diff --git a/dlls/shell32/shelllink.c b/dlls/shell32/shelllink.c
--- a/dlls/shell32/shelllink.c
+++ b/dlls/shell32/shelllink.c
@@ -62,7 +62,7 @@
 
 /* link file formats */
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 
 typedef struct _LINK_HEADER
 {
@@ -111,7 +111,7 @@
     WCHAR label[12];  /* assume 8.3 */
 } volume_info;
 
-#pragma pack(pop)
+#include "poppack.h"
 
 /* IShellLink Implementation */
 
