diff --git a/dlls/urlmon/urlmon.inf b/dlls/urlmon/urlmon.inf
--- a/dlls/urlmon/urlmon.inf
+++ b/dlls/urlmon/urlmon.inf
@@ -440,8 +440,6 @@
 ZONES_INTERNET="Software\Microsoft\Windows\CurrentVersion\Internet Settings\Zones\3"
 ZONES_UNTRUSTED="Software\Microsoft\Windows\CurrentVersion\Internet Settings\Zones\4"
 
-;; The English zone names and descriptions should be identical to the ones in urlmon.rc
-
 NAME_LOCAL="My Computer"
 DESC_LOCAL="Your computer"
 ZICON_LOCAL="explorer.exe#0100"
@@ -455,7 +453,7 @@
 ZICON_TRUSTED="inetcpl.cpl#00004480"
 
 NAME_INTERNET="Internet"
-DESC_INTERNET="This zone contains all Web sites you haven't placed in other zones."
+DESC_INTERNET="This zone contains all Web sites you haven't placed in other zones"
 ZICON_INTERNET="inetcpl.cpl#001313"
 
 NAME_UNTRUSTED="Restricted sites"
