diff --git a/dlls/windows.media.speech/classes.idl b/dlls/windows.media.speech/classes.idl
--- a/dlls/windows.media.speech/classes.idl
+++ b/dlls/windows.media.speech/classes.idl
@@ -15,15 +15,6 @@
  */
 
 #pragma makedep register
-#pragma winrt ns_prefix
 
-import "windows.media.speechrecognition.idl";
-import "windows.media.speechsynthesis.idl";
-
-namespace Windows.Media.SpeechRecognition {
-    runtimeclass SpeechRecognizer;
-    runtimeclass SpeechRecognitionListConstraint;
-}
-namespace Windows.Media.SpeechSynthesis {
-    runtimeclass SpeechSynthesizer;
-}
+#include "windows.media.speechrecognition.idl"
+#include "windows.media.speechsynthesis.idl"
