diff --git a/dlls/wined3d/query.c b/dlls/wined3d/query.c
--- a/dlls/wined3d/query.c
+++ b/dlls/wined3d/query.c
@@ -1915,7 +1915,7 @@
             break;
 
         case WINED3D_QUERY_TYPE_TIMESTAMP:
-            if (!wined3d_device_vk(device)->graphics_queue.timestamp_bits)
+            if (!wined3d_device_vk(device)->timestamp_bits)
             {
                 WARN("Timestamp queries not supported.\n");
                 return WINED3DERR_NOTAVAILABLE;
@@ -1925,7 +1925,7 @@
             break;
 
         case WINED3D_QUERY_TYPE_TIMESTAMP_DISJOINT:
-            if (!wined3d_device_vk(device)->graphics_queue.timestamp_bits)
+            if (!wined3d_device_vk(device)->timestamp_bits)
             {
                 WARN("Timestamp queries not supported.\n");
                 return WINED3DERR_NOTAVAILABLE;
