diff --git a/dlls/winsock.dll16/winsock16.h b/dlls/winsock.dll16/winsock16.h
--- a/dlls/winsock.dll16/winsock16.h
+++ b/dlls/winsock.dll16/winsock16.h
@@ -19,7 +19,7 @@
 #ifndef __WINE_WINE_WINSOCK16_H
 #define __WINE_WINE_WINSOCK16_H
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 typedef UINT16 SOCKET16;
 
@@ -74,7 +74,7 @@
     SEGPTR                  lpVendorInfo;
 } WSADATA16, *LPWSADATA16;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #define WS_FD_CLR16(fd, set)   __WS_FD_CLR((fd),(set), ws_fd_set16)
 #define WS_FD_SET16(fd, set)   __WS_FD_SET1((fd),(set), ws_fd_set16)
