diff --git a/dlls/wintrust/tests/softpub.c b/dlls/wintrust/tests/softpub.c
--- a/dlls/wintrust/tests/softpub.c
+++ b/dlls/wintrust/tests/softpub.c
@@ -656,7 +656,7 @@
 #define RVA_IDATA 0x3000
 #define FILE_TOTAL 0x600
 #define RVA_TOTAL 0x4000
-#pragma pack(push,1)
+#include <pshpack1.h>
 struct Imports {
     IMAGE_IMPORT_DESCRIPTOR descriptors[2];
     IMAGE_THUNK_DATA32 original_thunks[2];
@@ -735,7 +735,7 @@
     /* final alignment */
     {0}
 };
-#pragma pack(pop)
+#include <poppack.h>
 
 static void test_sip_create_indirect_data(void)
 {
