diff --git a/dlls/advapi32/Makefile.in b/dlls/advapi32/Makefile.in
--- a/dlls/advapi32/Makefile.in
+++ b/dlls/advapi32/Makefile.in
@@ -1,13 +1,16 @@
 EXTRADEFS = -D_ADVAPI32_
 MODULE    = advapi32.dll
 IMPORTLIB = advapi32
-IMPORTS   = kernelbase sechost msvcrt
-DELAYIMPORTS = cryptsp rpcrt4 user32 userenv
+IMPORTS   = kernelbase sechost msvcrt rpcrt4
+DELAYIMPORTS = user32
 
 SOURCES = \
 	advapi.c \
 	cred.c \
 	crypt.c \
+	crypt_arc4.c \
+	crypt_des.c \
+	crypt_lmhash.c \
 	eventlog.c \
 	lsa.c \
 	registry.c \
