diff --git a/dlls/comctl32/comctl32.rc b/dlls/comctl32/comctl32.rc
--- a/dlls/comctl32/comctl32.rc
+++ b/dlls/comctl32/comctl32.rc
@@ -111,6 +111,18 @@
 
 LANGUAGE LANG_NEUTRAL, SUBLANG_NEUTRAL
 
+#define WINE_FILEDESCRIPTION_STR "Wine Common Controls"
+#define WINE_FILENAME_STR "comctl32.dll"
+#define WINE_FILEVERSION COMCTL32_VERSION, COMCTL32_VERSION_MINOR, 4704, 1100
+#define WINE_FILEVERSION_STR "5.81"
+#define WINE_PRODUCTVERSION WINE_FILEVERSION
+#define WINE_PRODUCTVERSION_STR WINE_FILEVERSION_STR
+
+#include "wine/wine_common_ver.rc"
+
+/* @makedep: comctl32.manifest */
+WINE_MANIFEST RT_MANIFEST comctl32.manifest
+
 /* @makedep: idt_check.bmp */
 IDT_CHECK BITMAP idt_check.bmp
 
