diff --git a/dlls/d3dx11_43/Makefile.in b/dlls/d3dx11_43/Makefile.in
--- a/dlls/d3dx11_43/Makefile.in
+++ b/dlls/d3dx11_43/Makefile.in
@@ -1,7 +1,7 @@
-EXTRADEFS = -DD3DX11_SDK_VERSION=43
+EXTRADEFS = -DD3DX11_SDK_VERSION=43 -DD3DX_D3D_VERSION=11
 MODULE = d3dx11_43.dll
 IMPORTLIB = d3dx11
-IMPORTS   = d3dcompiler gdi32 ole32 user32
+IMPORTS   = d3dcompiler ole32
 PARENTSRC = ../d3dx9_36
 DELAYIMPORTS = windowscodecs
 
@@ -9,7 +9,7 @@
 
 SOURCES = \
 	async.c \
+	d3dx_helpers.c \
 	d3dx11_43_main.c \
-	d3dx_helpers.c \
 	texture.c \
 	version.rc
