diff --git a/dlls/hhctrl.ocx/hhctrl.rc b/dlls/hhctrl.ocx/hhctrl.rc
--- a/dlls/hhctrl.ocx/hhctrl.rc
+++ b/dlls/hhctrl.ocx/hhctrl.rc
@@ -34,8 +34,6 @@
 
     IDS_HIDETABS     "Hide &Tabs"
     IDS_SHOWTABS     "Show &Tabs"
-
-    IDS_SELECT_TOPIC "Select Topic"
 END
 
 STRINGTABLE
@@ -98,6 +96,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine htmlhelp OCX"
 #define WINE_FILENAME_STR "hhctrl.ocx"
 #define WINE_FILEVERSION 5,2,3790,2744
+#define WINE_FILEVERSION_STR "5.2.3790.2744"
+#define WINE_PRODUCTVERSION 5,2,3790,2744
+#define WINE_PRODUCTVERSION_STR "5.2.3790.2744"
 
 #include "wine/wine_common_ver.rc"
 
