diff --git a/dlls/msi/msi.c b/dlls/msi/msi.c
--- a/dlls/msi/msi.c
+++ b/dlls/msi/msi.c
@@ -2833,18 +2833,6 @@
     if (state == INSTALLSTATE_LOCAL && !*path)
         state = INSTALLSTATE_NOTUSED;
 
-    if (wcslen(path) > 2 && !wcsncmp(path, L"<\\", 2))
-    {
-        WCHAR *assembly_path = msi_get_assembly_path(path + 2);
-        if (assembly_path)
-        {
-            free(path);
-            path = assembly_path;
-            if (GetFileAttributesW(path) != INVALID_FILE_ATTRIBUTES)
-                state = INSTALLSTATE_LOCAL;
-        }
-    }
-
     if (msi_strcpy_to_awstring(path, -1, lpPathBuf, pcchBuf) == ERROR_MORE_DATA)
         state = INSTALLSTATE_MOREDATA;
 
