diff --git a/dlls/msvcp140_atomic_wait/main.c b/dlls/msvcp140_atomic_wait/main.c
--- a/dlls/msvcp140_atomic_wait/main.c
+++ b/dlls/msvcp140_atomic_wait/main.c
@@ -84,18 +84,6 @@
     return WaitOnAddress(addr, cmp, size, timeout);
 }
 
-void __stdcall __std_execution_wait_on_uchar(void *addr, unsigned char val)
-{
-    TRACE("(%p %d)\n", addr, val);
-    WaitOnAddress(addr, &val, sizeof(val), INFINITE);
-}
-
-void __stdcall __std_execution_wake_by_address_all(void *addr)
-{
-    TRACE("(%p)\n", addr);
-    WakeByAddressAll(addr);
-}
-
 typedef struct
 {
     SRWLOCK srwlock;
