diff --git a/dlls/windows.storage/classes.idl b/dlls/windows.storage/classes.idl
--- a/dlls/windows.storage/classes.idl
+++ b/dlls/windows.storage/classes.idl
@@ -19,16 +19,6 @@
  */
 
 #pragma makedep register
-#pragma winrt ns_prefix
 
-import "inspectable.idl";
-import "asyncinfo.idl";
-import "eventtoken.idl";
-import "windows.foundation.idl";
-import "windowscontracts.idl";
-import "windows.system.idl";
-import "windows.storage.idl";
-
-namespace Windows.Storage.Streams {
-    runtimeclass RandomAccessStreamReference;
-}
+#define _WINDOWS_STORAGE
+#include "windows.storage.streams.idl"
