diff --git a/dlls/winewayland.drv/display.c b/dlls/winewayland.drv/display.c
--- a/dlls/winewayland.drv/display.c
+++ b/dlls/winewayland.drv/display.c
@@ -193,7 +193,7 @@
 
     TRACE("\n");
 
-    device_manager->add_gpu(NULL, &pci_id, NULL, param);
+    device_manager->add_gpu("Wine GPU", &pci_id, NULL, param);
 }
 
 static void wayland_add_device_source(const struct gdi_device_manager *device_manager,
@@ -232,8 +232,7 @@
     mode->dmBitsPerPel = 32;
     mode->dmPelsWidth = output_mode->width;
     mode->dmPelsHeight = output_mode->height;
-    /* Round the refresh rate to calculate the win32 display frequency. */
-    mode->dmDisplayFrequency = (output_mode->refresh + 500) / 1000;
+    mode->dmDisplayFrequency = output_mode->refresh / 1000;
 }
 
 static void wayland_add_device_modes(const struct gdi_device_manager *device_manager,
