diff --git a/dlls/gdi32/uniscribe/indic.c b/dlls/gdi32/uniscribe/indic.c
--- a/dlls/gdi32/uniscribe/indic.c
+++ b/dlls/gdi32/uniscribe/indic.c
@@ -32,6 +32,7 @@
 #include "winternl.h"
 
 #include "wine/debug.h"
+#include "wine/heap.h"
 #include "usp10_internal.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(uniscribe);
@@ -346,7 +347,10 @@
         next = Indic_process_next_syllable(input, cChar, 0, &center, index, lex);
         if (next != -1)
         {
-            *syllables = realloc(*syllables, sizeof(IndicSyllable)*(*syllable_count+1));
+            if (*syllable_count)
+                *syllables = HeapReAlloc(GetProcessHeap(),0,*syllables, sizeof(IndicSyllable)*(*syllable_count+1));
+            else
+                *syllables = heap_alloc(sizeof(**syllables));
             (*syllables)[*syllable_count].start = index;
             (*syllables)[*syllable_count].base = center;
             (*syllables)[*syllable_count].ralf = -1;
