diff --git a/dlls/shell32/shell32.rc b/dlls/shell32/shell32.rc
--- a/dlls/shell32/shell32.rc
+++ b/dlls/shell32/shell32.rc
@@ -590,9 +590,6 @@
 /* @makedep: resources/drive.ico */
 IDI_SHELL_UNKNOWN_DRIVE ICON resources/drive.ico
 
-/* @makedep: resources/shield.ico */
-IDI_SHELL_SHIELD ICON resources/shield.ico
-
 /* FIXME: Following resource is not yet added */
 /* @makedep: resources/document.ico */
 IDI_SHELL_FILES ICON resources/document.ico
@@ -1119,6 +1116,9 @@
 
 #define WINE_FILENAME_STR "shell32.dll"
 #define WINE_FILEVERSION 6,0,2900,6242
+#define WINE_FILEVERSION_STR "6.0.2900.6242"
+#define WINE_PRODUCTVERSION 6,0,2900,6242
+#define WINE_PRODUCTVERSION_STR "6.0.2900.6242"
 #define WINE_EXTRAVALUES VALUE "OLESelfRegister",""
 
 #include "wine/wine_common_ver.rc"
