diff --git a/dlls/system.drv16/system.c b/dlls/system.drv16/system.c
--- a/dlls/system.drv16/system.c
+++ b/dlls/system.drv16/system.c
@@ -232,7 +232,9 @@
  */
 void WINAPI Save80x87State16( char *ptr )
 {
+#ifdef __i386__
     __asm__(".byte 0x66; fsave %0; fwait" : "=m" (ptr) );
+#endif
 }
 
 
@@ -241,7 +243,9 @@
  */
 void WINAPI Restore80x87State16( const char *ptr )
 {
+#ifdef __i386__
     __asm__(".byte 0x66; frstor %0" : : "m" (ptr) );
+#endif
 }
 
 
