diff --git a/dlls/windows.gaming.ui.gamebar/classes.idl b/dlls/windows.gaming.ui.gamebar/classes.idl
--- a/dlls/windows.gaming.ui.gamebar/classes.idl
+++ b/dlls/windows.gaming.ui.gamebar/classes.idl
@@ -19,14 +19,15 @@
  */
 
 #pragma makedep register
+
+#ifdef __WIDL__
 #pragma winrt ns_prefix
+#endif
 
 import "inspectable.idl";
 import "eventtoken.idl";
 import "windowscontracts.idl";
 import "windows.foundation.idl";
-import "windows.gaming.ui.idl";
 
-namespace Windows.Gaming.UI {
-    runtimeclass GameBar;
-}
+#define DO_NO_IMPORTS
+#include "windows.gaming.ui.idl"
