diff --git a/dlls/wineandroid.drv/android.h b/dlls/wineandroid.drv/android.h
--- a/dlls/wineandroid.drv/android.h
+++ b/dlls/wineandroid.drv/android.h
@@ -54,7 +54,9 @@
  */
 
 extern pthread_mutex_t drawable_mutex;
-extern UINT ANDROID_OpenGLInit( UINT version, const struct opengl_funcs *opengl_funcs, const struct opengl_driver_funcs **driver_funcs );
+extern void update_gl_drawable( HWND hwnd );
+extern void destroy_gl_drawable( HWND hwnd );
+extern struct opengl_funcs *get_wgl_driver( UINT version );
 
 
 /**************************************************************************
@@ -97,8 +99,6 @@
 extern BOOL ANDROID_CreateWindowSurface( HWND hwnd, BOOL layered, const RECT *surface_rect, struct window_surface **surface );
 extern void ANDROID_WindowPosChanged( HWND hwnd, HWND insert_after, HWND owner_hint, UINT swp_flags, BOOL fullscreen,
                                       const struct window_rects *new_rects, struct window_surface *surface );
-extern ANativeWindow *get_client_window( HWND hwnd );
-extern BOOL has_client_surface( HWND hwnd );
 
 /* unixlib interface */
 
