diff --git a/dlls/actxprxy/usrmarshal.c b/dlls/actxprxy/usrmarshal.c
--- a/dlls/actxprxy/usrmarshal.c
+++ b/dlls/actxprxy/usrmarshal.c
@@ -247,20 +247,3 @@
     TRACE("(%p)->(%p %p %p)\n", This, parent, folder, child);
     return IParentAndItem_RemoteGetParentAndItem_Proxy(This, parent, folder, child);
 }
-
-HRESULT CALLBACK IEnumObjects_Next_Proxy(IEnumObjects *This, ULONG celt, REFIID riid, void **rgelt, ULONG *pceltFetched)
-{
-    ULONG fetched;
-    TRACE("(%p)->(%ld, %p, %p, %p)\n", This, celt, debugstr_guid(riid), rgelt, pceltFetched);
-    if (!pceltFetched) pceltFetched = &fetched;
-    return IEnumObjects_RemoteNext_Proxy(This, celt, riid, rgelt, pceltFetched);
-}
-
-HRESULT __RPC_STUB IEnumObjects_Next_Stub(IEnumObjects *This, ULONG celt, REFIID riid, void **rgelt, ULONG *pceltFetched)
-{
-    HRESULT hr;
-    TRACE("(%p)->(%ld, %p, %p, %p)\n", This, celt, debugstr_guid(riid), rgelt, pceltFetched);
-    *pceltFetched = 0;
-    hr = IEnumObjects_Next(This, celt, riid, rgelt, pceltFetched);
-    return hr;
-}
