diff --git a/dlls/fusion/tests/asmname.c b/dlls/fusion/tests/asmname.c
--- a/dlls/fusion/tests/asmname.c
+++ b/dlls/fusion/tests/asmname.c
@@ -59,18 +59,9 @@
         return FALSE;
     }
 
-    hr = pLoadLibraryShim(L"fusion.dll", L"v4.0.30319", NULL, &hfusion);
+    hr = pLoadLibraryShim(L"fusion.dll", NULL, NULL, &hfusion);
     if (FAILED(hr))
-    {
-        hr = pLoadLibraryShim(L"fusion.dll", NULL, NULL, &hfusion);
-        if (FAILED(hr))
-        {
-            win_skip("fusion.dll not available\n");
-            FreeLibrary(hmscoree);
-            return FALSE;
-        }
-    }
-    else trace("using .NET version 4\n");
+        return FALSE;
 
     pCreateAssemblyNameObject = (void *)GetProcAddress(hfusion, "CreateAssemblyNameObject");
     if (!pCreateAssemblyNameObject)
