diff --git a/dlls/user32/winproc.c b/dlls/user32/winproc.c
--- a/dlls/user32/winproc.c
+++ b/dlls/user32/winproc.c
@@ -771,10 +771,7 @@
     case WM_COPYDATA:
     {
         COPYDATASTRUCT *cds = buffer;
-        /* If cbData <= 2048 bytes, the data is packed at the end of message buffer. Otherwise,
-         * cds->lpData points to an extra user buffer. See pack_user_message() for WM_COPYDATA */
-        if (cds->lpData && cds->cbData <= 2048)
-            cds->lpData = cds + 1;
+        if (cds->lpData) cds->lpData = cds + 1;
         break;
     }
     case EM_GETSEL:
@@ -807,7 +804,7 @@
     *lparam = (LPARAM)buffer;
 }
 
-NTSTATUS WINAPI User32CallWinProc( void *args, ULONG size )
+NTSTATUS WINAPI User32CallWindowProc( void *args, ULONG size )
 {
     struct win_proc_params *params = args;
     size_t packed_size = 0;
