diff --git a/dlls/comctl32/animate.c b/dlls/comctl32/animate.c
--- a/dlls/comctl32/animate.c
+++ b/dlls/comctl32/animate.c
@@ -899,11 +899,6 @@
     case WM_ERASEBKGND:
 	return ANIMATE_EraseBackground(infoPtr, (HDC)wParam);
 
-    case WM_GETOBJECT:
-        if ((LONG)lParam == OBJID_QUERYCLASSNAMEIDX)
-            return 0x1000e;
-	return DefWindowProcW(hWnd, uMsg, wParam, lParam);
-
     case WM_STYLECHANGED:
         return ANIMATE_StyleChanged(infoPtr, wParam, (LPSTYLESTRUCT)lParam);
 
@@ -971,3 +966,9 @@
 
     RegisterClassW(&wndClass);
 }
+
+
+void ANIMATE_Unregister(void)
+{
+    UnregisterClassW(ANIMATE_CLASSW, NULL);
+}
