diff --git a/dlls/gdi32/tests/bitmap.c b/dlls/gdi32/tests/bitmap.c
--- a/dlls/gdi32/tests/bitmap.c
+++ b/dlls/gdi32/tests/bitmap.c
@@ -5704,7 +5704,7 @@
     DWORD type, pixel;
     int size;
     HDC bmp_dc;
-    HBITMAP bmp, bmp2, tmp_bmp;
+    HBITMAP bmp;
 
     static const struct
     {
@@ -5893,12 +5893,6 @@
         bmp = SelectObject( bmp_dc, bmp );
         ret = BitBlt( bmp_dc, 0, 0, create_desc.Width, create_desc.Height, create_desc.hDc, 0, 0, SRCCOPY );
         ok(ret, "Failed to blit.\n");
-
-        /* cannot select a different bitmap on D3DKMT DCs */
-        bmp2 = CreateBitmap( 4, 4, 1, 32, NULL );
-        tmp_bmp = SelectObject( create_desc.hDc, bmp2 );
-        todo_wine ok( !tmp_bmp, "SelectObject succeeded\n" );
-        DeleteObject( bmp2 );
 
         destroy_desc.hDc = create_desc.hDc;
         destroy_desc.hBitmap = create_desc.hBitmap;
