diff --git a/dlls/joy.cpl/resource.h b/dlls/joy.cpl/resource.h
--- a/dlls/joy.cpl/resource.h
+++ b/dlls/joy.cpl/resource.h
@@ -37,7 +37,6 @@
 #define IDD_LIST            1000
 #define IDD_TEST_DI         1001
 #define IDD_TEST_XI         1002
-#define IDD_TEST_WGI        1003
 
 #define IDC_DI_ENABLED_LIST    2000
 #define IDC_XI_ENABLED_LIST    2001
@@ -70,14 +69,6 @@
 #define IDC_XI_RUMBLE_2     2222
 #define IDC_XI_RUMBLE_3     2223
 
-#define IDC_WGI_DEVICES     2300
-#define IDC_WGI_INTERFACE   2301
-#define IDC_WGI_GAMEPAD     2302
-#define IDC_WGI_RUMBLE      2303
-#define IDC_WGI_AXES        2304
-#define IDC_WGI_POVS        2305
-#define IDC_WGI_BUTTONS     2306
-
 #define ICO_MAIN            100
 
 #endif
