diff --git a/dlls/hid/tests/device.c b/dlls/hid/tests/device.c
--- a/dlls/hid/tests/device.c
+++ b/dlls/hid/tests/device.c
@@ -61,7 +61,7 @@
     status = HidP_GetLinkCollectionNodes(nodes, &nodes_count, ppd);
     ok(status == HIDP_STATUS_SUCCESS, "HidP_GetLinkCollectionNodes failed:%lx\n", status);
 
-    for (i = 0; i < min(nodes_count, ARRAY_SIZE(nodes)); ++i)
+    for (i = 0; i < nodes_count; ++i)
     {
         trace("  [%d] LinkUsage: %x LinkUsagePage: %x Parent: %x "
               "NumberOfChildren: %x NextSibling: %x FirstChild: %x "
@@ -370,13 +370,10 @@
         trace("REMAINING: %ld ms\n", max_time - spent);
     } while(spent < max_time);
 
-    CancelIo(device);
-    rc = WaitForSingleObject(overlapped.hEvent, timeout);
-    ok(rc == WAIT_OBJECT_0, "Wait for object failed\n");
-
     CloseHandle(overlapped.hEvent);
     rc = HidD_FreePreparsedData(ppd);
     ok(rc, "Failed to free preparsed data(0x%lx)\n", GetLastError());
+    CancelIo(device);
     CloseHandle(device);
     HeapFree(GetProcessHeap(), 0, data);
     HeapFree(GetProcessHeap(), 0, report);
