diff --git a/dlls/mf/samplegrabber.c b/dlls/mf/samplegrabber.c
--- a/dlls/mf/samplegrabber.c
+++ b/dlls/mf/samplegrabber.c
@@ -1170,9 +1170,9 @@
 {
     static const DWORD events[] =
     {
-        [SINK_STATE_STOPPED] = MEStreamSinkStopped,
-        [SINK_STATE_PAUSED]  = MEStreamSinkPaused,
-        [SINK_STATE_RUNNING] = MEStreamSinkStarted,
+        MEStreamSinkStopped, /* SINK_STATE_STOPPED */
+        MEStreamSinkPaused,  /* SINK_STATE_PAUSED */
+        MEStreamSinkStarted, /* SINK_STATE_RUNNING */
     };
     BOOL do_callback = FALSE;
     HRESULT hr = S_OK;
@@ -1195,6 +1195,10 @@
 
             if (state == SINK_STATE_RUNNING && grabber->state != SINK_STATE_RUNNING)
             {
+                /* Unpause at a position is a seek operation which drops everything pending. */
+                if (grabber->state == SINK_STATE_PAUSED && offset != PRESENTATION_CURRENT_POSITION)
+                    grabber->sample_count = MAX_SAMPLE_QUEUE_LENGTH;
+
                 /* Every transition to running state sends a bunch requests to build up initial queue. */
                 for (i = 0; i < grabber->sample_count; ++i)
                 {
