diff --git a/dlls/secur32/lsa.c b/dlls/secur32/lsa.c
--- a/dlls/secur32/lsa.c
+++ b/dlls/secur32/lsa.c
@@ -768,14 +768,6 @@
         return SEC_E_UNSUPPORTED_FUNCTION;
 
     return lsa_ctx->package->user_api->VerifySignature(lsa_ctx->handle, message, message_seq_no, quality_of_protection);
-}
-
-static SECURITY_STATUS WINAPI lsa_QuerySecurityContextToken(CtxtHandle *context, HANDLE *token)
-{
-    FIXME("%p %p): stub\n", context, token);
-    if (!OpenProcessToken(GetCurrentProcess(), MAXIMUM_ALLOWED, token))
-        return GetLastError();
-    return SEC_E_OK;
 }
 
 static SECURITY_STATUS WINAPI lsa_EncryptMessage(CtxtHandle *context, ULONG quality_of_protection,
@@ -838,7 +830,7 @@
     NULL, /* ImportSecurityContextW */
     NULL, /* AddCredentialsW */
     NULL, /* Reserved8 */
-    lsa_QuerySecurityContextToken,
+    NULL, /* QuerySecurityContextToken */
     lsa_EncryptMessage,
     lsa_DecryptMessage,
     NULL, /* SetContextAttributesW */
