diff --git a/dlls/secur32/secur32.c b/dlls/secur32/secur32.c
--- a/dlls/secur32/secur32.c
+++ b/dlls/secur32/secur32.c
@@ -1158,9 +1158,22 @@
             return FALSE;
         }
 
+    case NameDisplay:
+        {
+            static const WCHAR wineusernameW[] = {'W','I','N','E','U','S','E','R','N','A','M','E',0};
+
+            DWORD needed = GetEnvironmentVariableW(wineusernameW, NULL, 0);
+            if (*nSize < needed) {
+                *nSize = needed;
+                SetLastError(ERROR_MORE_DATA);
+                return FALSE;
+            }
+            *nSize = GetEnvironmentVariableW(wineusernameW, lpNameBuffer, *nSize);
+            return TRUE;
+        }
+
     case NameUnknown:
     case NameFullyQualifiedDN:
-    case NameDisplay:
     case NameUniqueId:
     case NameCanonical:
     case NameUserPrincipal:
