diff --git a/dlls/wineps.drv/bitmap.c b/dlls/wineps.drv/bitmap.c
--- a/dlls/wineps.drv/bitmap.c
+++ b/dlls/wineps.drv/bitmap.c
@@ -108,10 +108,20 @@
        the foregnd color corresponds to a bit equal to
        0 in the bitmap.
     */
-    PSDRV_CreateColor(ctx, &foregnd, RGB(info->bmiColors[0].rgbRed, info->bmiColors[0].rgbGreen,
-                      info->bmiColors[0].rgbBlue));
-    PSDRV_CreateColor(ctx, &bkgnd, RGB(info->bmiColors[1].rgbRed, info->bmiColors[1].rgbGreen,
-                      info->bmiColors[1].rgbBlue));
+    if (!info->bmiHeader.biClrUsed)
+    {
+        PSDRV_CreateColor( ctx, &foregnd, GetTextColor( ctx->hdc ) );
+        bkgnd = ctx->bkColor;
+    }
+    else
+    {
+        PSDRV_CreateColor( ctx, &foregnd, RGB(info->bmiColors[0].rgbRed,
+                                              info->bmiColors[0].rgbGreen,
+                                              info->bmiColors[0].rgbBlue) );
+        PSDRV_CreateColor( ctx, &bkgnd, RGB(info->bmiColors[1].rgbRed,
+                                            info->bmiColors[1].rgbGreen,
+                                            info->bmiColors[1].rgbBlue) );
+    }
 
     PSDRV_WriteGSave(ctx);
     PSDRV_WriteNewPath(ctx);
