diff --git a/dlls/combase/tests/wine.combase.test.c b/dlls/combase/tests/wine.combase.test.c
--- a/dlls/combase/tests/wine.combase.test.c
+++ b/dlls/combase/tests/wine.combase.test.c
@@ -100,7 +100,7 @@
 {
     struct factory *impl = impl_from_IActivationFactory(iface);
 
-    TRACE("iface %p, trust_level %p.\n", iface, trust_level);
+    FIXME("iface %p, trust_level %p stub!\n", iface, trust_level);
 
     if (!impl->trusted) return E_NOTIMPL;
 
@@ -132,6 +132,7 @@
 
 HRESULT WINAPI DllCanUnloadNow(void)
 {
+    FIXME("stub!\n");
     return S_OK;
 }
 
@@ -139,7 +140,7 @@
 {
     const WCHAR *buffer = WindowsGetStringRawBuffer(classid, NULL);
 
-    TRACE("class %s, factory %p.\n", debugstr_w(buffer), factory);
+    FIXME("class %s, factory %p stub!\n", debugstr_w(buffer), factory);
 
     if (!wcscmp(buffer, L"Wine.Test.Class"))
     {
