diff --git a/dlls/msvcrt/misc.c b/dlls/msvcrt/misc.c
--- a/dlls/msvcrt/misc.c
+++ b/dlls/msvcrt/misc.c
@@ -212,6 +212,8 @@
  *  the wrong calling convention or parameters.
  */
 #ifdef __i386__
+
+# if defined(__GNUC__) || defined(__clang__)
 
 __ASM_GLOBAL_FUNC(_chkesp,
                   "jnz 1f\n\t"
@@ -240,6 +242,16 @@
   DebugBreak();
 }
 
+# else  /* __GNUC__ || __clang__ */
+
+/**********************************************************************/
+
+void CDECL _chkesp(void)
+{
+}
+
+# endif  /* __GNUC__ || __clang__ */
+
 #endif  /* __i386__ */
 
 static inline void swap(char *l, char *r, size_t size)
