diff --git a/dlls/msxml3/comment.c b/dlls/msxml3/comment.c
--- a/dlls/msxml3/comment.c
+++ b/dlls/msxml3/comment.c
@@ -158,9 +158,11 @@
 {
     domcomment *This = impl_from_IXMLDOMComment( iface );
 
+    static const WCHAR commentW[] = {'#','c','o','m','m','e','n','t',0};
+
     TRACE("(%p)->(%p)\n", This, p);
 
-    return return_bstr(L"#comment", p);
+    return return_bstr(commentW, p);
 }
 
 static HRESULT WINAPI domcomment_get_nodeValue(
@@ -349,10 +351,11 @@
     BSTR* p)
 {
     domcomment *This = impl_from_IXMLDOMComment( iface );
+    static const WCHAR commentW[] = {'c','o','m','m','e','n','t',0};
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return return_bstr(L"comment", p);
+    return return_bstr(commentW, p);
 }
 
 static HRESULT WINAPI domcomment_get_text(
