diff --git a/dlls/opencl/make_opencl b/dlls/opencl/make_opencl
--- a/dlls/opencl/make_opencl
+++ b/dlls/opencl/make_opencl
@@ -589,11 +589,6 @@
 
 #include "config.h"
 #include "unix_private.h"
-
-#ifdef __APPLE__
-#pragma clang diagnostic push
-#pragma clang diagnostic ignored "-Wdeprecated-declarations"
-#endif
 EOF
 ;
 
@@ -604,13 +599,6 @@
     print UNIX "\n", generate_unix_thunk( $_, $core_functions{$_} );
 }
 
-print UNIX <<EOF
-#ifdef __APPLE__
-#pragma clang diagnostic pop
-#endif
-EOF
-;
-
 print UNIX "\nconst unixlib_entry_t __wine_unix_call_funcs[] =\n{\n";
 foreach (sort keys %core_functions)
 {
