diff --git a/dlls/shell32/shlview.c b/dlls/shell32/shlview.c
--- a/dlls/shell32/shlview.c
+++ b/dlls/shell32/shlview.c
@@ -458,7 +458,7 @@
 
         lvColumn.fmt = sd.fmt;
         lvColumn.cx = MulDiv(sd.cxChar, tm.tmAveCharWidth * 3, 2); /* chars->pixel */
-        StrRetToBufW(&sd.str, NULL, nameW, ARRAY_SIZE(nameW));
+        StrRetToStrNW(nameW, ARRAY_SIZE(nameW), &sd.str, NULL);
         SendMessageW(This->hWndList, LVM_INSERTCOLUMNW, This->columns, (LPARAM)&lvColumn);
     }
 
@@ -1424,14 +1424,14 @@
 
               if (lpnmh->code == LVN_GETDISPINFOW)
               {
-                  StrRetToBufW(&sd.str, NULL, lpdi->item.pszText, lpdi->item.cchTextMax);
+                  StrRetToStrNW( lpdi->item.pszText, lpdi->item.cchTextMax, &sd.str, NULL);
                   TRACE("-- text=%s\n", debugstr_w(lpdi->item.pszText));
               }
               else
               {
                   /* LVN_GETDISPINFOA - shouldn't happen */
                   NMLVDISPINFOA *lpdiA = (NMLVDISPINFOA *)lpnmh;
-                  StrRetToBufA(&sd.str, NULL, lpdiA->item.pszText, lpdiA->item.cchTextMax);
+                  StrRetToStrNA( lpdiA->item.pszText, lpdiA->item.cchTextMax, &sd.str, NULL);
                   TRACE("-- text=%s\n", lpdiA->item.pszText);
               }
 	    }
