diff --git a/dlls/win32u/painting.c b/dlls/win32u/painting.c
--- a/dlls/win32u/painting.c
+++ b/dlls/win32u/painting.c
@@ -489,16 +489,6 @@
 {
     FIXME( "(%p): stub\n", hdc );
     return FALSE;
-}
-
-
-/***********************************************************************
- *           NtGdiCancelDC   (win32u.@)
- */
-BOOL WINAPI NtGdiCancelDC( HDC hdc )
-{
-    FIXME( "%p: stub\n", hdc );
-    return TRUE;
 }
 
 
@@ -922,7 +912,7 @@
  */
 BOOL WINAPI NtGdiDrawStream( HDC hdc, ULONG in, void *pvin )
 {
-    FIXME("stub: %p, %d, %p\n", hdc, in, pvin);
+    FIXME("stub: %p, %d, %p\n", hdc, (int)in, pvin);
     return FALSE;
 }
 
