diff --git a/dlls/winegstreamer/Makefile.in b/dlls/winegstreamer/Makefile.in
--- a/dlls/winegstreamer/Makefile.in
+++ b/dlls/winegstreamer/Makefile.in
@@ -7,9 +7,16 @@
 UNIX_LIBS    = $(GSTREAMER_LIBS) $(PTHREAD_LIBS)
 
 SOURCES = \
-	aac_decoder.c \
+	audio_decoder.c \
 	color_convert.c \
 	main.c \
+	media-converter/audioconv.c \
+	media-converter/audioconvbin.c \
+	media-converter/fossilize.c \
+	media-converter/lib.c \
+	media-converter/murmur3.c \
+	media-converter/protondemuxer.c \
+	media-converter/videoconv.c \
 	media_sink.c \
 	media_source.c \
 	mfplat.c \
@@ -27,6 +34,7 @@
 	wg_muxer.c \
 	wg_parser.c \
 	wg_sample.c \
+	wg_task_pool.c \
 	wg_transform.c \
 	winegstreamer_classes.idl \
 	wm_reader.c \
