diff --git a/dlls/winemac.drv/cocoa_clipboard.m b/dlls/winemac.drv/cocoa_clipboard.m
--- a/dlls/winemac.drv/cocoa_clipboard.m
+++ b/dlls/winemac.drv/cocoa_clipboard.m
@@ -46,9 +46,9 @@
 /***********************************************************************
  *              macdrv_is_pasteboard_owner
  */
-bool macdrv_is_pasteboard_owner(macdrv_window w)
-{
-    __block bool ret;
+int macdrv_is_pasteboard_owner(macdrv_window w)
+{
+    __block int ret;
     WineWindow* window = (WineWindow*)w;
 
     OnMainThread(^{
@@ -65,10 +65,10 @@
 /***********************************************************************
  *              macdrv_has_pasteboard_changed
  */
-bool macdrv_has_pasteboard_changed(void)
+int macdrv_has_pasteboard_changed(void)
 {
     __block int new_change_count;
-    bool ret;
+    int ret;
 
     OnMainThread(^{
         NSPasteboard* pb = [NSPasteboard generalPasteboard];
