diff --git a/dlls/wmp/oleobj.c b/dlls/wmp/oleobj.c
--- a/dlls/wmp/oleobj.c
+++ b/dlls/wmp/oleobj.c
@@ -157,7 +157,7 @@
         IOleInPlaceSiteEx_OnInPlaceActivateEx(ipsiteex, &redraw, 0);
         ipsite = (IOleInPlaceSite*)ipsiteex;
     }else {
-        hres = IOleClientSite_QueryInterface(This->client_site, &IID_IOleInPlaceSite, (void**)&ipsite);
+        IOleClientSite_QueryInterface(This->client_site, &IID_IOleInPlaceSite, (void**)&ipsite);
         if(FAILED(hres)) {
             FIXME("No IOleInPlaceSite instance\n");
             return hres;
@@ -421,15 +421,7 @@
 
         ShowWindow(This->hwnd, SW_HIDE);
         return S_OK;
-    case OLEIVERB_SHOW:
-        if (!This->hwnd)
-        {
-            FIXME("No window to show\n");
-            return E_UNEXPECTED;
-        }
-
-        ShowWindow(This->hwnd, SW_SHOW);
-        return S_OK;
+
     default:
         FIXME("Unsupported iVerb %ld\n", iVerb);
     }
