diff --git a/dlls/jscript/jscript.rc b/dlls/jscript/jscript.rc
--- a/dlls/jscript/jscript.rc
+++ b/dlls/jscript/jscript.rc
@@ -51,7 +51,6 @@
     IDS_DISABLED_CC         "Conditional compilation is turned off"
     IDS_EXPECTED_AT         "Expected '@'"
     IDS_NOT_FUNC            "Function expected"
-    IDS_NOT_STRING          "'[object]' is not a string object"
     IDS_NOT_DATE            "'[object]' is not a date object"
     IDS_NOT_NUM             "Number expected"
     IDS_OBJECT_EXPECTED     "Object expected"
@@ -70,7 +69,6 @@
     IDS_URI_INVALID_CHAR    "URI to be encoded contains invalid characters"
     IDS_FRACTION_DIGITS_OUT_OF_RANGE "Number of fraction digits is out of range"
     IDS_PRECISION_OUT_OF_RANGE "Precision is out of range"
-    IDS_ARRAY_OR_ARGS_EXPECTED "Array or arguments object expected"
     IDS_INVALID_LENGTH      "Array length must be a finite positive integer"
     IDS_ARRAY_EXPECTED      "Array object expected"
     IDS_INVALID_WRITABLE_PROP_DESC "'writable' attribute on the property descriptor cannot be set to 'true' on this object"
@@ -105,6 +103,9 @@
 
 #define WINE_FILEDESCRIPTION_STR "Wine JScript"
 #define WINE_FILENAME_STR "jscript.dll"
-#define WINE_FILEVERSION 5,8,7601,16982
+#define WINE_FILEVERSION 5,7,7601,16982
+#define WINE_FILEVERSION_STR "5.8.7601.16982"
+#define WINE_PRODUCTVERSION 5,8,7601,16982
+#define WINE_PRODUCTVERSION_STR "5.8.7601.16982"
 
 #include "wine/wine_common_ver.rc"
