diff --git a/dlls/krnl386.exe16/ne_module.c b/dlls/krnl386.exe16/ne_module.c
--- a/dlls/krnl386.exe16/ne_module.c
+++ b/dlls/krnl386.exe16/ne_module.c
@@ -42,7 +42,7 @@
 WINE_DECLARE_DEBUG_CHANNEL(relay);
 WINE_DECLARE_DEBUG_CHANNEL(winediag);
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 typedef struct _GPHANDLERDEF
 {
     WORD selector;
@@ -50,7 +50,7 @@
     WORD rangeEnd;
     WORD handler;
 } GPHANDLERDEF;
-#pragma pack(pop)
+#include "poppack.h"
 
 /*
  * Segment table entry
@@ -915,7 +915,7 @@
     }
 
     patch_code_segment( pModule );
-    *(ULONG *)mz_header->e_res2 = NtCurrentTeb()->Peb->SpareUlongs[0];
+    *(const void **)mz_header->e_res2 = ldt_copy->base;
 
     return hInstance;
 }
