diff --git a/dlls/shell32/recyclebin.c b/dlls/shell32/recyclebin.c
--- a/dlls/shell32/recyclebin.c
+++ b/dlls/shell32/recyclebin.c
@@ -104,7 +104,7 @@
     }
     else
     {
-        const WCHAR *data_home = _wgetenv( L"WINE_HOST_XDG_DATA_HOME" );
+        const WCHAR *data_home = _wgetenv( L"XDG_DATA_HOME" );
         const WCHAR *fmt = L"%s/.local/share/Trash";
         WCHAR *p;
 
@@ -207,7 +207,7 @@
     sprintf( buffer + strlen(buffer), "\nDeletionDate=%04u-%02u-%02uT%02u:%02u:%02u\n",
              now.wYear, now.wMonth, now.wDay, now.wHour, now.wMinute, now.wSecond);
     WriteFile( handle, buffer, strlen(buffer), NULL, NULL );
-    HeapFree( GetProcessHeap(), 0, path );
+    heap_free( path );
     free( buffer );
     return TRUE;
 }
@@ -240,7 +240,7 @@
             if ((path = wine_get_dos_file_name( decode( p ))))
             {
                 lstrcpynW( data->cFileName, path, MAX_PATH );
-                HeapFree( GetProcessHeap(), 0, path );
+                heap_free( path );
             }
             else
             {
