diff --git a/dlls/storage.dll16/storage.c b/dlls/storage.dll16/storage.c
--- a/dlls/storage.dll16/storage.c
+++ b/dlls/storage.dll16/storage.c
@@ -341,7 +341,7 @@
     assert(n>=-1);
     if (str->hf) {
 	if ((SetFilePointer( str->hf, (n+1)*BIGSIZE, NULL,
-			     FILE_BEGIN ) == INVALID_SET_FILE_POINTER) && GetLastError())
+			     SEEK_SET ) == INVALID_SET_FILE_POINTER) && GetLastError())
 	{
             WARN("(%p,%d,%p), seek failed (%ld)\n",str->hf, n, block, GetLastError());
 	    return FALSE;
@@ -422,7 +422,7 @@
     assert(n>=-1);
     if (str->hf) {
 	if ((SetFilePointer( str->hf, (n+1)*BIGSIZE, NULL,
-			     FILE_BEGIN ) == INVALID_SET_FILE_POINTER) && GetLastError())
+			     SEEK_SET ) == INVALID_SET_FILE_POINTER) && GetLastError())
 	{
             WARN("seek failed (%ld)\n",GetLastError());
 	    return FALSE;
@@ -738,7 +738,7 @@
         DWORD result;
 
 	if (str->hf)
-	    SetFilePointer( str->hf, 0, NULL, FILE_BEGIN );
+	    SetFilePointer( str->hf, 0, NULL, SEEK_SET );
 	/* block -1 is the storage header */
 	sth = (struct storage_header*)block;
 	memcpy(sth->magic,STORAGE_magic,8);
