diff --git a/dlls/win32u/clipboard.c b/dlls/win32u/clipboard.c
--- a/dlls/win32u/clipboard.c
+++ b/dlls/win32u/clipboard.c
@@ -347,8 +347,24 @@
  */
 INT WINAPI NtUserGetClipboardFormatName( UINT format, WCHAR *buffer, INT maxlen )
 {
-    UNICODE_STRING str = {.Buffer = buffer, .MaximumLength = maxlen};
-    return NtUserGetAtomName( format, &str );
+    char buf[sizeof(ATOM_BASIC_INFORMATION) + MAX_ATOM_LEN * sizeof(WCHAR)];
+    ATOM_BASIC_INFORMATION *abi = (ATOM_BASIC_INFORMATION *)buf;
+    UINT length = 0;
+
+    if (format < MAXINTATOM || format > 0xffff) return 0;
+    if (maxlen <= 0)
+    {
+        RtlSetLastWin32Error( ERROR_MORE_DATA );
+        return 0;
+    }
+    if (!set_ntstatus( NtQueryInformationAtom( format, AtomBasicInformation,
+                                               buf, sizeof(buf), NULL )))
+        return 0;
+
+    length = min( abi->NameLength / sizeof(WCHAR), maxlen - 1 );
+    if (length) memcpy( buffer, abi->Name, length * sizeof(WCHAR) );
+    buffer[length] = 0;
+    return length;
 }
 
 /**************************************************************************
