diff --git a/dlls/win32u/region.c b/dlls/win32u/region.c
--- a/dlls/win32u/region.c
+++ b/dlls/win32u/region.c
@@ -830,7 +830,7 @@
     DWORD size;
     WINEREGION *obj = GDI_GetObjPtr( hrgn, NTGDI_OBJ_REGION );
 
-    TRACE(" %p count = %d, rgndata = %p\n", hrgn, count, rgndata);
+    TRACE(" %p count = %d, rgndata = %p\n", hrgn, (int)count, rgndata);
 
     if(!obj) return 0;
 
@@ -901,7 +901,7 @@
 
     /* XP doesn't care about the type */
     if( rgndata->rdh.iType != RDH_RECTANGLES )
-        WARN("(Unsupported region data type: %u)\n", rgndata->rdh.iType);
+        WARN("(Unsupported region data type: %u)\n", (int)rgndata->rdh.iType);
 
     if (xform)
     {
@@ -948,7 +948,7 @@
 done:
     if (!hrgn) free_region( obj );
 
-    TRACE("%p %d %p returning %p\n", xform, count, rgndata, hrgn );
+    TRACE("%p %d %p returning %p\n", xform, (int)count, rgndata, hrgn );
     return hrgn;
 }
 
