diff --git a/dlls/dwrite/scripts.h b/dlls/dwrite/scripts.h
--- a/dlls/dwrite/scripts.h
+++ b/dlls/dwrite/scripts.h
@@ -1,5 +1,5 @@
 /* Unicode Script IDs */
-/* generated from https://www.unicode.org/Public/17.0.0/ucd/UCD.zip:Scripts.txt */
+/* generated from https://www.unicode.org/Public/16.0.0/ucd/UCD.zip:Scripts.txt */
 /* DO NOT EDIT!! */
 
 enum unicode_script_id {
@@ -160,19 +160,5 @@
     Script_Dives_Akuru = 154,
     Script_Khitan_Small_Script = 155,
     Script_Yezidi = 156,
-    Script_Cypro_Minoan = 157,
-    Script_Kawi = 158,
-    Script_Nag_Mundari = 159,
-    Script_Old_Uyghur = 160,
-    Script_Tangsa = 161,
-    Script_Toto = 162,
-    Script_Vithkuqi = 163,
-    Script_Garay = 164,
-    Script_Gurung_Khema = 165,
-    Script_Kirat_Rai = 166,
-    Script_Ol_Onal = 167,
-    Script_Sunuwar = 168,
-    Script_Todhri = 169,
-    Script_Tulu_Tigalari = 170,
-    Script_LastId = 170
+    Script_LastId = 156
 };
