diff --git a/dlls/ole32/ole2.c b/dlls/ole32/ole2.c
--- a/dlls/ole32/ole2.c
+++ b/dlls/ole32/ole2.c
@@ -733,7 +733,6 @@
   TrackerWindowInfo trackerInfo;
   HWND            hwndTrackWindow;
   MSG             msg;
-  HCURSOR         cursor;
 
   TRACE("%p, %p, %#lx, %p.\n", pDataObject, pDropSource, dwOKEffect, pdwEffect);
 
@@ -767,9 +766,6 @@
     SetCapture(hwndTrackWindow);
 
     msg.message = 0;
-
-    /* save cursor */
-    cursor = GetCursor();
 
     /*
      * Pump messages. All mouse input should go to the capture window.
@@ -816,9 +812,6 @@
      * Destroy the temporary window.
      */
     DestroyWindow(hwndTrackWindow);
-
-    /* restore cursor */
-    SetCursor(cursor);
 
     return trackerInfo.returnValue;
   }
