diff --git a/dlls/win32u/Makefile.in b/dlls/win32u/Makefile.in
--- a/dlls/win32u/Makefile.in
+++ b/dlls/win32u/Makefile.in
@@ -2,9 +2,9 @@
 MODULE    = win32u.dll
 UNIXLIB   = win32u.so
 IMPORTLIB = win32u
-IMPORTS   = ntdll
+IMPORTS   = ntdll winecrt0
 UNIX_CFLAGS  = $(FREETYPE_CFLAGS) $(FONTCONFIG_CFLAGS)
-UNIX_LIBS    = $(CORETEXT_LIBS) $(APPKIT_LIBS) $(PTHREAD_LIBS) -lm
+UNIX_LIBS    = $(CARBON_LIBS) $(APPKIT_LIBS) $(PTHREAD_LIBS) -lm
 
 EXTRADLLFLAGS = -nodefaultlibs
 
@@ -25,6 +25,7 @@
 	dibdrv/dc.c \
 	dibdrv/graphics.c \
 	dibdrv/objects.c \
+	dibdrv/opengl.c \
 	dibdrv/primitives.c \
 	driver.c \
 	emfdrv.c \
@@ -38,7 +39,6 @@
 	mapping.c \
 	menu.c \
 	message.c \
-	opengl.c \
 	opentype.c \
 	painting.c \
 	palette.c \
