diff --git a/dlls/d3dx11_43/d3dx11_43.spec b/dlls/d3dx11_43/d3dx11_43.spec
--- a/dlls/d3dx11_43/d3dx11_43.spec
+++ b/dlls/d3dx11_43/d3dx11_43.spec
@@ -12,20 +12,20 @@
 @ stdcall D3DX11CreateAsyncResourceLoaderA(long str ptr)
 @ stdcall D3DX11CreateAsyncResourceLoaderW(long wstr ptr)
 @ stub D3DX11CreateAsyncShaderPreprocessProcessor
-@ stub D3DX11CreateAsyncShaderResourceViewProcessor
-@ stub D3DX11CreateAsyncTextureInfoProcessor
-@ stub D3DX11CreateAsyncTextureProcessor
-@ stub D3DX11CreateShaderResourceViewFromFileA
-@ stub D3DX11CreateShaderResourceViewFromFileW
+@ stdcall D3DX11CreateAsyncShaderResourceViewProcessor(ptr ptr ptr)
+@ stdcall D3DX11CreateAsyncTextureInfoProcessor(ptr ptr)
+@ stdcall D3DX11CreateAsyncTextureProcessor(ptr ptr ptr)
+@ stdcall D3DX11CreateShaderResourceViewFromFileA(ptr str ptr ptr ptr ptr)
+@ stdcall D3DX11CreateShaderResourceViewFromFileW(ptr wstr ptr ptr ptr ptr)
 @ stdcall D3DX11CreateShaderResourceViewFromMemory(ptr ptr long ptr ptr ptr ptr)
-@ stub D3DX11CreateShaderResourceViewFromResourceA
-@ stub D3DX11CreateShaderResourceViewFromResourceW
+@ stdcall D3DX11CreateShaderResourceViewFromResourceA(ptr long str ptr ptr ptr ptr)
+@ stdcall D3DX11CreateShaderResourceViewFromResourceW(ptr long wstr ptr ptr ptr ptr)
 @ stdcall D3DX11CreateTextureFromFileA(ptr str ptr ptr ptr ptr)
 @ stdcall D3DX11CreateTextureFromFileW(ptr wstr ptr ptr ptr ptr)
 @ stdcall D3DX11CreateTextureFromMemory(ptr ptr long ptr ptr ptr ptr)
 @ stdcall D3DX11CreateTextureFromResourceA(ptr long str ptr ptr ptr ptr)
 @ stdcall D3DX11CreateTextureFromResourceW(ptr long wstr ptr ptr ptr ptr)
-@ stub D3DX11CreateThreadPump
+@ stdcall D3DX11CreateThreadPump(long long ptr)
 @ stdcall D3DX11FilterTexture(ptr ptr long long)
 @ stdcall D3DX11GetImageInfoFromFileA(str ptr ptr ptr)
 @ stdcall D3DX11GetImageInfoFromFileW(wstr ptr ptr ptr)
