diff --git a/dlls/krnl386.exe16/int26.c b/dlls/krnl386.exe16/int26.c
--- a/dlls/krnl386.exe16/int26.c
+++ b/dlls/krnl386.exe16/int26.c
@@ -48,7 +48,7 @@
                     0, NULL);
     if (h != INVALID_HANDLE_VALUE)
     {
-        SetFilePointer(h, begin * 512, NULL, FILE_BEGIN );
+        SetFilePointer(h, begin * 512, NULL, SEEK_SET );
         /* FIXME: check errors */
         WriteFile( h, dataptr, nr_sect * 512, &w, NULL );
         CloseHandle( h );
@@ -68,7 +68,7 @@
 void WINAPI DOSVM_Int26Handler( CONTEXT *context )
 {
     WCHAR drivespec[] = {'A', ':', '\\', 0};
-    BYTE *dataptr = ldt_get_ptr( context->SegDs, context->Ebx );
+    BYTE *dataptr = CTX_SEG_OFF_TO_LIN( context, context->SegDs, context->Ebx );
     DWORD begin;
     DWORD length;
 
@@ -86,7 +86,9 @@
     {
         begin   = *(DWORD *)dataptr;
         length  = *(WORD *)(dataptr + 4);
-        dataptr = ldt_get_ptr( *(WORD *)(dataptr + 8), *(DWORD *)(dataptr + 6) );
+        dataptr = (BYTE *)CTX_SEG_OFF_TO_LIN( context,
+                                              *(WORD *)(dataptr + 8), 
+                                              *(DWORD *)(dataptr + 6) );
     }
     else
     {
