diff --git a/dlls/dsound/dsound_main.c b/dlls/dsound/dsound_main.c
--- a/dlls/dsound/dsound_main.c
+++ b/dlls/dsound/dsound_main.c
@@ -79,6 +79,8 @@
 GUID *DSOUND_renderer_guids;
 GUID *DSOUND_capture_guids;
 
+const WCHAR wine_vxd_drv[] = L"winemm.vxd";
+
 /* All default settings, you most likely don't want to touch these, see wiki on UsefulRegistryKeys */
 int ds_hel_buflen = 32768 * 2;
 
@@ -406,7 +408,6 @@
     PROPVARIANT pv;
     BOOL keep_going;
     HRESULT hr;
-    WCHAR *id;
 
     PropVariantInit(&pv);
 
@@ -433,10 +434,8 @@
     TRACE("Calling back with %s (%s)\n", wine_dbgstr_guid(guid),
             wine_dbgstr_w(pv.pwszVal));
 
-    hr = IMMDevice_GetId(device, &id);
-    keep_going = cb(guid, pv.pwszVal, id, user);
-
-    CoTaskMemFree(id);
+    keep_going = cb(guid, pv.pwszVal, wine_vxd_drv, user);
+
     PropVariantClear(&pv);
     IPropertyStore_Release(ps);
 
