diff --git a/dlls/kernel32/tests/comm.c b/dlls/kernel32/tests/comm.c
--- a/dlls/kernel32/tests/comm.c
+++ b/dlls/kernel32/tests/comm.c
@@ -1977,12 +1977,9 @@
                 test_GetModemStatus(hcom);
                 break;
             }
-            else if (last_event_time || after - before >= 10000)
+            else
             {
-                CancelIoEx(hcom, &ovl_wait);
-                ret = WaitForSingleObject(ovl_wait.hEvent, 500);
-                ok(ret == WAIT_OBJECT_0, "Wait for object failed\n");
-                goto done;
+                if (last_event_time || after - before >= 10000) goto done;
             }
         }
     } while (after - before < 10000);
@@ -2186,12 +2183,9 @@
                 ok(iob.Information == 1, "expected 1, got %Iu\n", iob.Information);
                 goto done;
             }
-            else if (last_event_time || after - before >= 3000)
+            else
             {
-                CancelIoEx(hcom, &ovl_wait);
-                ret = WaitForSingleObject(ovl_wait.hEvent, 100);
-                ok(ret == WAIT_OBJECT_0, "Wait object failed\n");
-                goto done;
+                if (last_event_time || after - before >= 3000) goto done;
             }
         }
     } while (after - before < 3000);
