diff --git a/dlls/xinput9_1_0/main.c b/dlls/xinput9_1_0/main.c
--- a/dlls/xinput9_1_0/main.c
+++ b/dlls/xinput9_1_0/main.c
@@ -70,12 +70,27 @@
     return TRUE;
 }
 
-DWORD WINAPI DECLSPEC_HOTPATCH XInputGetCapabilities(DWORD index, DWORD flags, XINPUT_CAPABILITIES *capabilities)
+DWORD WINAPI DECLSPEC_HOTPATCH XInputGetCapabilities(DWORD index, DWORD flags, XINPUT_CAPABILITIES *caps)
 {
+    DWORD ret;
+
     InitOnceExecuteOnce(&init_xinput1_4_once, init_xinput1_4_funcs, NULL, NULL);
 
     if (!pXInputGetCapabilities) return ERROR_DEVICE_NOT_CONNECTED;
-    return pXInputGetCapabilities(index, flags, capabilities);
+    if (!(ret = pXInputGetCapabilities(index, flags, caps)))
+    {
+        caps->Flags = XINPUT_CAPS_VOICE_SUPPORTED;
+        caps->Gamepad.bLeftTrigger = !!caps->Gamepad.bLeftTrigger;
+        caps->Gamepad.bRightTrigger = !!caps->Gamepad.bRightTrigger;
+        caps->Gamepad.sThumbLX = !!caps->Gamepad.sThumbLX;
+        caps->Gamepad.sThumbLY = !!caps->Gamepad.sThumbLY;
+        caps->Gamepad.sThumbRX = !!caps->Gamepad.sThumbRX;
+        caps->Gamepad.sThumbRY = !!caps->Gamepad.sThumbRY;
+        caps->Vibration.wLeftMotorSpeed = !!caps->Vibration.wLeftMotorSpeed;
+        caps->Vibration.wRightMotorSpeed = !!caps->Vibration.wRightMotorSpeed;
+    }
+
+    return ret;
 }
 
 DWORD WINAPI DECLSPEC_HOTPATCH XInputGetDSoundAudioDeviceGuids(DWORD index, GUID *render_guid, GUID *capture_guid)
