diff --git a/dlls/jscript/jscript.c b/dlls/jscript/jscript.c
--- a/dlls/jscript/jscript.c
+++ b/dlls/jscript/jscript.c
@@ -912,6 +912,17 @@
             WARN("object does not implement IDispatch\n");
             return hres;
         }
+
+        if(!cc_api.note_edge && This->ctx->html_mode) {
+            IWineJSDispatchHost *host;
+
+            /* Init CC API to get rid of cycles in IE8 and below as well */
+            hres = IDispatch_QueryInterface(disp, &IID_IWineJSDispatchHost, (void**)&host);
+            if(SUCCEEDED(hres) && host) {
+                init_cc_api(host);
+                IWineJSDispatchHost_Release(host);
+            }
+        }
     }
 
     item = malloc(sizeof(*item));
@@ -1459,10 +1470,10 @@
 }
 
 static HRESULT WINAPI WineJScript_InitHostConstructor(IWineJScript *iface, IWineJSDispatchHost *constr,
-                                                      const WCHAR *method_name, IWineJSDispatch **ret)
+                                                      IWineJSDispatch *prototype, IWineJSDispatch **ret)
 {
     JScript *This = impl_from_IWineJScript(iface);
-    return init_host_constructor(This->ctx, constr, method_name, ret);
+    return init_host_constructor(This->ctx, constr, prototype, ret);
 }
 
 static HRESULT WINAPI WineJScript_CreateObject(IWineJScript *iface, IWineJSDispatch **ret)
