diff --git a/dlls/msi/msipriv.h b/dlls/msi/msipriv.h
--- a/dlls/msi/msipriv.h
+++ b/dlls/msi/msipriv.h
@@ -428,6 +428,16 @@
     UINT   LastActionResult;
     UINT   action_progress_increment;
     HANDLE log_file;
+    HMODULE hfusion10;
+    HMODULE hfusion11;
+    HMODULE hfusion20;
+    HMODULE hfusion40;
+    HMODULE hmscoree;
+    HRESULT (WINAPI *pGetFileVersion)( const WCHAR *, WCHAR *, DWORD, DWORD * );
+    HRESULT (WINAPI *pCreateAssemblyNameObject)( IAssemblyName **, const WCHAR *, DWORD, void * );
+    HRESULT (WINAPI *pCreateAssemblyEnum)( IAssemblyEnum **, IUnknown *, IAssemblyName *, DWORD, void * );
+    IAssemblyCache *cache_net[CLR_VERSION_MAX];
+    IAssemblyCache *cache_sxs;
 
     struct list classes;
     struct list extensions;
@@ -1076,10 +1086,10 @@
 extern MSIASSEMBLY *msi_load_assembly(MSIPACKAGE *, MSICOMPONENT *);
 extern UINT msi_install_assembly(MSIPACKAGE *, MSICOMPONENT *);
 extern UINT msi_uninstall_assembly(MSIPACKAGE *, MSICOMPONENT *);
-extern void msi_destroy_assembly_caches(void);
+extern void msi_destroy_assembly_caches(MSIPACKAGE *);
 extern BOOL msi_is_global_assembly(MSICOMPONENT *);
-extern IAssemblyEnum *msi_create_assembly_enum(const WCHAR *);
-extern WCHAR *msi_get_assembly_path(const WCHAR *) __WINE_DEALLOC(free) __WINE_MALLOC;
+extern IAssemblyEnum *msi_create_assembly_enum(MSIPACKAGE *, const WCHAR *);
+extern WCHAR *msi_get_assembly_path(MSIPACKAGE *, const WCHAR *) __WINE_DEALLOC(free) __WINE_MALLOC;
 extern WCHAR **msi_split_string(const WCHAR *, WCHAR);
 extern UINT msi_set_original_database_property(MSIDATABASE *, const WCHAR *);
 extern WCHAR *msi_get_error_message(MSIDATABASE *, int) __WINE_DEALLOC(free) __WINE_MALLOC;
