diff --git a/dlls/qedit/main.c b/dlls/qedit/main.c
--- a/dlls/qedit/main.c
+++ b/dlls/qedit/main.c
@@ -16,6 +16,8 @@
  * License along with this library; if not, write to the Free Software
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
+
+#define WINE_NO_NAMELESS_EXTENSION
 
 #include "qedit_private.h"
 #include "rpcproxy.h"
@@ -205,8 +207,8 @@
 {
     .dwVersion = 2,
     .dwMerit = MERIT_DO_NOT_USE,
-    .cPins2 = 2,
-    .rgPins2 = reg_sample_grabber_pins,
+    .u.s2.cPins2 = 2,
+    .u.s2.rgPins2 = reg_sample_grabber_pins,
 };
 
 static const REGFILTERPINS2 reg_null_renderer_pins[1] =
@@ -223,8 +225,8 @@
 {
     .dwVersion = 2,
     .dwMerit = MERIT_DO_NOT_USE,
-    .cPins2 = 1,
-    .rgPins2 = reg_null_renderer_pins,
+    .u.s2.cPins2 = 1,
+    .u.s2.rgPins2 = reg_null_renderer_pins,
 };
 
 /***********************************************************************
