diff --git a/dlls/urlmon/mimefilter.c b/dlls/urlmon/mimefilter.c
--- a/dlls/urlmon/mimefilter.c
+++ b/dlls/urlmon/mimefilter.c
@@ -544,13 +544,12 @@
         if(!proposed_mime)
             return E_FAIL;
 
-        len = wcscspn(proposed_mime, L";");
-        *ret_mime = CoTaskMemAlloc((len+1)*sizeof(WCHAR));
+        len = lstrlenW(proposed_mime)+1;
+        *ret_mime = CoTaskMemAlloc(len*sizeof(WCHAR));
         if(!*ret_mime)
             return E_OUTOFMEMORY;
 
         memcpy(*ret_mime, proposed_mime, len*sizeof(WCHAR));
-        (*ret_mime)[len] = 0;
         return S_OK;
     }
 
@@ -639,13 +638,12 @@
 
     TRACE("found %s for %s\n", debugstr_w(ret), debugstr_an((const char*)buf, min(32, size)));
 
-    len = wcscspn(ret, L";");
-    *ret_mime = CoTaskMemAlloc((len+1)*sizeof(WCHAR));
+    len = lstrlenW(ret)+1;
+    *ret_mime = CoTaskMemAlloc(len*sizeof(WCHAR));
     if(!*ret_mime)
         return E_OUTOFMEMORY;
 
     memcpy(*ret_mime, ret, len*sizeof(WCHAR));
-    (*ret_mime)[len] = 0;
     return S_OK;
 }
 
