diff --git a/dlls/winegstreamer/unixlib.h b/dlls/winegstreamer/unixlib.h
--- a/dlls/winegstreamer/unixlib.h
+++ b/dlls/winegstreamer/unixlib.h
@@ -145,6 +145,7 @@
             uint32_t payload_type;
             uint32_t codec_data_len;
             unsigned char codec_data[64];
+            UINT8 is_xma;
         } audio;
 
         /* Valid members for different video formats:
@@ -192,6 +193,7 @@
     UINT32 flags;
     UINT32 max_size;
     UINT32 size;
+    UINT32 stride;
     UINT64 data; /* pointer to user memory */
 };
 
@@ -221,6 +223,7 @@
 {
     wg_parser_t parser;
     UINT8 output_compressed;
+    UINT8 use_opengl;
     UINT8 err_on;
     UINT8 warn_on;
 };
@@ -334,6 +337,7 @@
 struct wg_transform_attrs
 {
     UINT32 output_plane_align;
+    UINT32 output_plane_stride;
     UINT32 input_queue_length;
     BOOL allow_format_change;
     BOOL low_latency;
