diff --git a/dlls/dmusic/soundfont.h b/dlls/dmusic/soundfont.h
--- a/dlls/dmusic/soundfont.h
+++ b/dlls/dmusic/soundfont.h
@@ -24,7 +24,7 @@
 
 #include "wine/debug.h"
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 /* SoundFont 2.04 data structures, from http://www.synthfont.com/sfspec24.pdf */
 
@@ -277,14 +277,6 @@
     }
 }
 
-enum
-{
-    SF_UNLOOPED = 0,
-    SF_LOOP_DURING_RELEASE = 1,
-    SF_NOTUSED = 2,
-    SF_LOOP_UNTIL_RELEASE = 3
-};
-
 struct sf_instrument /* <inst-rec> */
 {
     char name[20];
@@ -309,7 +301,7 @@
 
 C_ASSERT(sizeof(struct sf_sample) == 46);
 
-#pragma pack(pop)
+#include <poppack.h>
 
 struct soundfont
 {
