diff --git a/dlls/mscoree/mscoree_main.c b/dlls/mscoree/mscoree_main.c
--- a/dlls/mscoree/mscoree_main.c
+++ b/dlls/mscoree/mscoree_main.c
@@ -800,7 +800,7 @@
     return ret;
 }
 
-static BOOL get_support_msi(LPCWSTR mono_path, LPWSTR msi_path)
+static BOOL get_support_msi(LPCWSTR mono_path, LPWSTR msi_path, BOOL ignore_version)
 {
     static const WCHAR support_msi_relative[] = {'\\','s','u','p','p','o','r','t','\\','w','i','n','e','m','o','n','o','-','s','u','p','p','o','r','t','.','m','s','i',0};
     UINT (WINAPI *pMsiOpenPackageW)(LPCWSTR,ULONG*);
@@ -838,7 +838,7 @@
     if (res == ERROR_SUCCESS) {
         TRACE("found support msi version %s at %s\n", versionstringbuf, debugstr_w(msi_path));
 
-        if (compare_versions(WINE_MONO_VERSION, versionstringbuf) <= 0)
+        if (ignore_version || compare_versions(WINE_MONO_VERSION, versionstringbuf) <= 0)
         {
             ret = TRUE;
         }
@@ -922,12 +922,12 @@
 
     initresult = CoInitialize(NULL);
 
-    ret = get_support_msi(mono_path, support_msi_path);
+    ret = get_support_msi(mono_path, support_msi_path, FALSE);
     if (!ret)
     {
         /* Try looking outside c:\windows\mono */
         ret = (get_mono_path(mono_path, TRUE) &&
-            get_support_msi(mono_path, support_msi_path));
+            get_support_msi(mono_path, support_msi_path, TRUE));
     }
 
     if (ret)
