diff --git a/dlls/shell32/shlexec.c b/dlls/shell32/shlexec.c
--- a/dlls/shell32/shlexec.c
+++ b/dlls/shell32/shlexec.c
@@ -49,6 +49,7 @@
 typedef UINT_PTR (*SHELL_ExecuteW32)(const WCHAR *lpCmd, WCHAR *env, BOOL shWait,
 			    const SHELLEXECUTEINFOW *sei, LPSHELLEXECUTEINFOW sei_out);
 extern BOOL WINAPI PathResolveW(void *path, const WCHAR **paths, DWORD flags);
+extern BOOL WINAPI PathFileExistsDefExtW(LPWSTR lpszPath,DWORD dwWhich);
 
 static inline BOOL isSpace(WCHAR c)
 {
@@ -286,7 +287,7 @@
 	hr = IShellFolder_GetDisplayNameOf(desktop, pidl, SHGDN_FORPARSING, &strret);
 
 	if (SUCCEEDED(hr))
-	    StrRetToBufW(&strret, pidl, pszPath, uOutSize);
+	    StrRetToStrNW(pszPath, uOutSize, &strret, pidl);
 
 	IShellFolder_Release(desktop);
     }
@@ -1206,13 +1207,13 @@
         if (r != ERROR_SUCCESS )
             return hkey;
 
-        r = RegQueryValueExW( hkey, NULL, 0, &type, NULL, &sz );
+        r = RegQueryValueExW( hkey, L"", 0, &type, NULL, &sz );
         if ( r == ERROR_SUCCESS && type == REG_SZ )
         {
             sz += sizeof (WCHAR);
             cls = malloc( sz );
             cls[0] = 0;
-            RegQueryValueExW( hkey, NULL, 0, &type, (LPBYTE) cls, &sz );
+            RegQueryValueExW( hkey, L"", 0, &type, (LPBYTE) cls, &sz );
         }
 
         RegCloseKey( hkey );
