diff --git a/dlls/msftedit/msftedit_main.c b/dlls/msftedit/msftedit_main.c
--- a/dlls/msftedit/msftedit_main.c
+++ b/dlls/msftedit/msftedit_main.c
@@ -28,9 +28,14 @@
 #include "winuser.h"
 #include "richedit.h"
 #include "imm.h"
+#include "shlwapi.h"
 #include "oleidl.h"
 #include "initguid.h"
 #include "textserv.h"
+
+#include "wine/debug.h"
+
+WINE_DEFAULT_DEBUG_CHANNEL(msftedit);
 
 /***********************************************************************
  * DllMain.
@@ -52,3 +57,18 @@
     }
     return TRUE;
 }
+
+/***********************************************************************
+ *              DllGetVersion (msftedit.@)
+ */
+HRESULT WINAPI DllGetVersion(DLLVERSIONINFO *info)
+{
+    if (info->cbSize != sizeof(DLLVERSIONINFO)) FIXME("support DLLVERSIONINFO2\n");
+
+    /* this is what WINXP SP2 reports */
+    info->dwMajorVersion = 41;
+    info->dwMinorVersion = 15;
+    info->dwBuildNumber = 1507;
+    info->dwPlatformID = 1;
+    return NOERROR;
+}
