diff --git a/dlls/crypt32/tests/cert.c b/dlls/crypt32/tests/cert.c
--- a/dlls/crypt32/tests/cert.c
+++ b/dlls/crypt32/tests/cert.c
@@ -275,12 +275,10 @@
             CertFreeCertificateContext(context);
         }
 
-        ret = CertCloseStore(collection, CERT_CLOSE_STORE_CHECK_FLAG);
-        ok(ret, "got error %#lx.\n", GetLastError());
+        CertCloseStore(collection, 0);
     }
 
-    ret = CertCloseStore(store, CERT_CLOSE_STORE_CHECK_FLAG);
-    ok(ret, "got error %#lx.\n", GetLastError());
+    CertCloseStore(store, 0);
 }
 
 static void checkHash(const BYTE *data, DWORD dataLen, ALG_ID algID,
@@ -4418,8 +4416,9 @@
 
     ret = CertDeleteCertificateFromStore(cert);
     ok(ret, "CertDeleteCertificateFromStore error %#lx\n", GetLastError());
-    ret = CertCloseStore(store, CERT_CLOSE_STORE_CHECK_FLAG);
-    ok(ret, "got error %#lx.\n", GetLastError());
+
+    CertFreeCertificateContext(cert);
+    CertCloseStore(store, 0);
 }
 
 static void test_VerifySignature(void)
@@ -4519,6 +4518,7 @@
     testGetIssuerCert();
     testLinkCert();
     testKeyProvInfo();
+
     testCryptHashCert();
     testCryptHashCert2();
     testCertSigs();
