diff --git a/dlls/hhctrl.ocx/hhctrl.h b/dlls/hhctrl.ocx/hhctrl.h
--- a/dlls/hhctrl.ocx/hhctrl.h
+++ b/dlls/hhctrl.ocx/hhctrl.h
@@ -134,7 +134,6 @@
     HWND hwndEdit;
     HWND hwndList;
     HWND hwndContainer;
-    HWND hwndSearchBtn;
 } SearchTab;
 
 typedef struct {
@@ -169,17 +168,6 @@
     char *pszCustomTabs;
 };
 
-struct HHInfo;
-
-typedef struct WebBrowserEvents2Impl
-{
-    IDispatch WebBrowserEvents2Impl_iface;
-
-    struct HHInfo *info;
-    LONG ref;
-    DWORD cookie;
-} WebBrowserEvents2Impl;
-
 typedef struct {
     IOleClientSite IOleClientSite_iface;
     IOleInPlaceSite IOleInPlaceSite_iface;
@@ -191,8 +179,6 @@
     IOleObject *ole_obj;
     IWebBrowser2 *web_browser;
     HWND hwndWindow;
-
-    WebBrowserEvents2Impl *WebBrowser_events_sink;
 } WebBrowserContainer;
 
 typedef struct {
