diff --git a/dlls/shell32/autocomplete.c b/dlls/shell32/autocomplete.c
--- a/dlls/shell32/autocomplete.c
+++ b/dlls/shell32/autocomplete.c
@@ -205,7 +205,7 @@
 {
     UINT cur = 0, array_size = 1024;
     LPOLESTR *strs = NULL, *tmp;
-    BOOL str_read = FALSE;
+    ULONG read;
 
     do
     {
@@ -215,20 +215,15 @@
 
         do
         {
-            /* An implementation of IEnumString::Next in the ETQW World Editor
-               never initializes the output number of strings returned, so to
-               determine if a string was successfully retrieved, just enumerate
-               one string at a time and check the returned HRESULT. */
-            ULONG dummy;
-            HRESULT hr = IEnumString_Next(ac->enumstr, 1, &strs[cur], &dummy);
-            str_read = SUCCEEDED(hr) && hr != S_FALSE;
-        } while (str_read && ++cur < array_size);
+            if (FAILED(IEnumString_Next(ac->enumstr, array_size - cur, &strs[cur], &read)))
+                read = 0;
+        } while (read != 0 && (cur += read) < array_size);
 
         array_size *= 2;
-    } while (str_read);
+    } while (read != 0);
 
     /* Allocate even if there were zero strings enumerated, to mark it non-NULL */
-    if ((tmp = realloc(strs, max(cur, 1) * sizeof(*strs))))
+    if ((tmp = realloc(strs, cur * sizeof(*strs))))
     {
         strs = tmp;
         if (cur > 0)
