diff --git a/dlls/msxml3/selection.c b/dlls/msxml3/selection.c
--- a/dlls/msxml3/selection.c
+++ b/dlls/msxml3/selection.c
@@ -169,8 +169,7 @@
     if ( ref == 0 )
     {
         xmlXPathFreeObject(This->result);
-        if (This->node)
-            xmldoc_release(This->node->doc);
+        xmldoc_release(This->node->doc);
         if (This->enumvariant) IEnumVARIANT_Release(This->enumvariant);
         free(This);
     }
@@ -834,25 +833,3 @@
     xmlXPathFreeContext(ctxt);
     return hr;
 }
-
-HRESULT create_selection_from_nodeset(xmlXPathObjectPtr nodeset, IXMLDOMNodeList **out)
-{
-    domselection *obj;
-
-    TRACE("%p, %p.\n", nodeset, out);
-
-    *out = NULL;
-
-    if (!(obj = calloc(1, sizeof(*obj))))
-        return E_OUTOFMEMORY;
-
-    obj->IXMLDOMSelection_iface.lpVtbl = &domselection_vtbl;
-    obj->ref = 1;
-    init_dispex(&obj->dispex, (IUnknown *)&obj->IXMLDOMSelection_iface, &domselection_dispex);
-
-    obj->result = nodeset;
-
-    *out = (IXMLDOMNodeList *)&obj->IXMLDOMSelection_iface;
-
-    return S_OK;
-}
