diff --git a/dlls/krnl386.exe16/interrupts.c b/dlls/krnl386.exe16/interrupts.c
--- a/dlls/krnl386.exe16/interrupts.c
+++ b/dlls/krnl386.exe16/interrupts.c
@@ -148,7 +148,9 @@
 {
     if (islong)
     {
-        DWORD *stack = ldt_get_ptr(context->SegSs, context->Esp);
+        DWORD *stack = CTX_SEG_OFF_TO_LIN(context, 
+                                          context->SegSs, 
+                                          context->Esp);
         context->Esp += -4; /* One item will be added to stack. */
 
         if (isstub)
@@ -161,11 +163,13 @@
             *(--stack) = ip;
         }
         else
-            *(--stack) = context->EFlags;
+            *(--stack) = context->EFlags;            
     }
     else
     {
-        WORD *stack = ldt_get_ptr(context->SegSs, context->Esp);
+        WORD *stack = CTX_SEG_OFF_TO_LIN(context, 
+                                         context->SegSs, 
+                                         context->Esp);
         ADD_LOWORD( context->Esp, -2 ); /* One item will be added to stack. */
 
         if (isstub)
@@ -249,7 +253,8 @@
     if (context->SegCs == int16_sel)
     {
         /* Restore original flags stored into the stack by the caller. */
-        WORD *stack = ldt_get_ptr(context->SegSs, context->Esp);
+        WORD *stack = CTX_SEG_OFF_TO_LIN(context, 
+                                         context->SegSs, context->Esp);
         context->EFlags = (DWORD)MAKELONG( stack[2], HIWORD(context->EFlags) );
 
         if (intnum != context->Eip / DOSVM_STUB_PM16)
