diff --git a/dlls/windows.gaming.input/classes.idl b/dlls/windows.gaming.input/classes.idl
--- a/dlls/windows.gaming.input/classes.idl
+++ b/dlls/windows.gaming.input/classes.idl
@@ -19,7 +19,10 @@
  */
 
 #pragma makedep register
+
+#ifdef __WIDL__
 #pragma winrt ns_prefix
+#endif
 
 import "inspectable.idl";
 import "asyncinfo.idl";
@@ -30,21 +33,8 @@
 import "windows.devices.haptics.idl";
 import "windows.system.idl";
 import "windows.devices.power.idl";
-import "windows.gaming.input.idl";
-import "windows.gaming.input.custom.idl";
-import "windows.gaming.input.forcefeedback.idl";
 
-namespace Windows.Gaming.Input {
-    runtimeclass RawGameController;
-    runtimeclass Gamepad;
-    runtimeclass RacingWheel;
-}
-namespace Windows.Gaming.Input.Custom {
-    runtimeclass GameControllerFactoryManager;
-}
-namespace Windows.Gaming.Input.ForceFeedback {
-    runtimeclass ConstantForceEffect;
-    runtimeclass RampForceEffect;
-    runtimeclass PeriodicForceEffect;
-    runtimeclass ConditionForceEffect;
-}
+#define DO_NO_IMPORTS
+#include "windows.gaming.input.forcefeedback.idl"
+#include "windows.gaming.input.idl"
+#include "windows.gaming.input.custom.idl"
