diff --git a/dlls/dwrite/gdiinterop.c b/dlls/dwrite/gdiinterop.c
--- a/dlls/dwrite/gdiinterop.c
+++ b/dlls/dwrite/gdiinterop.c
@@ -693,8 +693,8 @@
 static HRESULT WINAPI gdiinterop_ConvertFontToLOGFONT(IDWriteGdiInterop1 *iface,
     IDWriteFont *font, LOGFONTW *logfont, BOOL *is_systemfont)
 {
-    IDWriteFontFileLoader *loader;
-    IDWriteFontFile *file;
+    IDWriteFontCollection *collection;
+    IDWriteFontFamily *family;
     HRESULT hr;
 
     TRACE("%p, %p, %p, %p.\n", iface, font, logfont, is_systemfont);
@@ -703,16 +703,20 @@
 
     memset(logfont, 0, sizeof(*logfont));
 
-    file = get_fontfile_from_font(font);
-
-    if (!font || !file)
+    if (!font)
         return E_INVALIDARG;
 
-    if (FAILED(hr = IDWriteFontFile_GetLoader(file, &loader)))
+    hr = IDWriteFont_GetFontFamily(font, &family);
+    if (FAILED(hr))
         return hr;
 
-    *is_systemfont = loader == get_local_fontfile_loader();
-    IDWriteFontFileLoader_Release(loader);
+    hr = IDWriteFontFamily_GetFontCollection(family, &collection);
+    IDWriteFontFamily_Release(family);
+    if (FAILED(hr))
+        return hr;
+
+    *is_systemfont = is_system_collection(collection);
+    IDWriteFontCollection_Release(collection);
 
     get_logfont_from_font(font, logfont);
     logfont->lfCharSet = DEFAULT_CHARSET;
