diff --git a/dlls/explorerframe/taskbarlist.c b/dlls/explorerframe/taskbarlist.c
--- a/dlls/explorerframe/taskbarlist.c
+++ b/dlls/explorerframe/taskbarlist.c
@@ -170,18 +170,18 @@
 
 static HRESULT STDMETHODCALLTYPE taskbar_list_UnregisterTab(ITaskbarList4 *iface, HWND hwndTab)
 {
-    FIXME("iface %p, hwndTab %p stub, faking success!\n", iface, hwndTab);
-
-    return S_OK;
+    FIXME("iface %p, hwndTab %p stub!\n", iface, hwndTab);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE taskbar_list_SetTabOrder(ITaskbarList4 *iface,
                                                           HWND hwndTab,
                                                           HWND hwndInsertBefore)
 {
-    FIXME("iface %p, hwndTab %p, hwndInsertBefore %p stub, faking success!\n", iface, hwndTab, hwndInsertBefore);
-
-    return S_OK;
+    FIXME("iface %p, hwndTab %p, hwndInsertBefore %p stub!\n", iface, hwndTab, hwndInsertBefore);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE taskbar_list_SetTabActive(ITaskbarList4 *iface,
@@ -189,9 +189,9 @@
                                                           HWND hwndMDI,
                                                           DWORD dwReserved)
 {
-    FIXME("iface %p, hwndTab %p, hwndMDI %p, dwReserved %lx stub, faking success!\n", iface, hwndTab, hwndMDI, dwReserved);
-
-    return S_OK;
+    FIXME("iface %p, hwndTab %p, hwndMDI %p, dwReserved %lx stub!\n", iface, hwndTab, hwndMDI, dwReserved);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE taskbar_list_ThumbBarAddButtons(ITaskbarList4 *iface,
