diff --git a/dlls/gdiplus/gdiplus_private.h b/dlls/gdiplus/gdiplus_private.h
--- a/dlls/gdiplus/gdiplus_private.h
+++ b/dlls/gdiplus/gdiplus_private.h
@@ -42,7 +42,7 @@
 #define VERSION_MAGIC  0xdbc01001
 #define VERSION_MAGIC2 0xdbc01002
 #define VALID_MAGIC(x) (((x) & 0xfffff000) == 0xdbc01000)
-#define TENSION_CONST (0.333333333f)
+#define TENSION_CONST (0.3)
 
 #define GIF_DISPOSE_UNSPECIFIED 0
 #define GIF_DISPOSE_DO_NOT_DISPOSE 1
@@ -149,8 +149,6 @@
 typedef struct region_element region_element;
 extern void delete_element(region_element *element);
 
-extern GpStatus get_region_hrgn(struct region_element *element, const RECT *bounds, HRGN *hrgn);
-
 extern GpStatus get_hatch_data(GpHatchStyle hatchstyle, const unsigned char **result);
 
 static inline INT gdip_round(REAL x)
@@ -398,19 +396,6 @@
 
 typedef struct CGpEffect{
     EffectType type;
-    union {
-        BYTE data[1];
-        struct BlurParams blur;
-        struct TintParams tint;
-        struct RedEyeCorrectionParams redeye;
-        ColorMatrix matrix;
-        struct ColorLUTParams lut;
-        struct BrightnessContrastParams brightness;
-        struct HueSaturationLightnessParams hue;
-        struct ColorBalanceParams balance;
-        struct LevelsParams levels;
-        struct ColorCurveParams curve;
-    } params;
 } CGpEffect;
 
 struct GpImage{
