diff --git a/dlls/d3d11/async.c b/dlls/d3d11/async.c
--- a/dlls/d3d11/async.c
+++ b/dlls/d3d11/async.c
@@ -71,7 +71,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(query->device);
+        ID3D11Device2_AddRef(query->device);
         wined3d_query_incref(query->wined3d_query);
     }
 
@@ -87,9 +87,9 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = query->device;
+        ID3D11Device2 *device = query->device;
         wined3d_query_decref(query->wined3d_query);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -245,7 +245,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(query->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(query->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_query_GetPrivateData(ID3D10Query *iface,
@@ -457,7 +457,7 @@
     wined3d_mutex_unlock();
 
     query->predicate = predicate;
-    ID3D11Device5_AddRef(query->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(query->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
