diff --git a/dlls/d3dx10_43/d3dx10_43.spec b/dlls/d3dx10_43/d3dx10_43.spec
--- a/dlls/d3dx10_43/d3dx10_43.spec
+++ b/dlls/d3dx10_43/d3dx10_43.spec
@@ -15,7 +15,7 @@
 @ stdcall D3DX10CreateAsyncResourceLoaderA(long str ptr)
 @ stdcall D3DX10CreateAsyncResourceLoaderW(long wstr ptr)
 @ stub D3DX10CreateAsyncShaderPreprocessProcessor(str ptr ptr ptr ptr ptr)
-@ stub D3DX10CreateAsyncShaderResourceViewProcessor(ptr ptr ptr)
+@ stdcall D3DX10CreateAsyncShaderResourceViewProcessor(ptr ptr ptr)
 @ stdcall D3DX10CreateAsyncTextureInfoProcessor(ptr ptr)
 @ stdcall D3DX10CreateAsyncTextureProcessor(ptr ptr ptr)
 @ stdcall D3DX10CreateDevice(ptr long long long ptr)
@@ -35,11 +35,11 @@
 @ stdcall D3DX10CreateFontIndirectW(ptr ptr ptr)
 @ stdcall D3DX10CreateFontW(ptr long long long long long long long long long wstr ptr)
 @ stdcall D3DX10CreateMesh(ptr ptr long str long long long ptr)
-@ stub D3DX10CreateShaderResourceViewFromFileA(ptr str ptr ptr ptr ptr)
-@ stub D3DX10CreateShaderResourceViewFromFileW(ptr wstr ptr ptr ptr ptr)
-@ stub D3DX10CreateShaderResourceViewFromMemory(ptr ptr long ptr ptr ptr ptr)
-@ stub D3DX10CreateShaderResourceViewFromResourceA(ptr long str ptr ptr ptr ptr)
-@ stub D3DX10CreateShaderResourceViewFromResourceW(ptr long wstr ptr ptr ptr ptr)
+@ stdcall D3DX10CreateShaderResourceViewFromFileA(ptr str ptr ptr ptr ptr)
+@ stdcall D3DX10CreateShaderResourceViewFromFileW(ptr wstr ptr ptr ptr ptr)
+@ stdcall D3DX10CreateShaderResourceViewFromMemory(ptr ptr long ptr ptr ptr ptr)
+@ stdcall D3DX10CreateShaderResourceViewFromResourceA(ptr long str ptr ptr ptr ptr)
+@ stdcall D3DX10CreateShaderResourceViewFromResourceW(ptr long wstr ptr ptr ptr ptr)
 @ stub D3DX10CreateSkinInfo(ptr)
 @ stdcall D3DX10CreateSprite(ptr long ptr)
 @ stdcall D3DX10CreateTextureFromFileA(ptr str ptr ptr ptr ptr)
@@ -61,9 +61,9 @@
 @ stub D3DX10PreprocessShaderFromResourceA(long str str ptr ptr ptr ptr ptr)
 @ stub D3DX10PreprocessShaderFromResourceW(long wstr wstr ptr ptr ptr ptr ptr)
 @ stub D3DX10SHProjectCubeMap(long ptr ptr ptr ptr)
-@ stub D3DX10SaveTextureToFileA(ptr ptr str)
-@ stub D3DX10SaveTextureToFileW(ptr ptr wstr)
-@ stub D3DX10SaveTextureToMemory(ptr ptr ptr long)
+@ stdcall D3DX10SaveTextureToFileA(ptr ptr str)
+@ stdcall D3DX10SaveTextureToFileW(ptr ptr wstr)
+@ stdcall D3DX10SaveTextureToMemory(ptr ptr ptr long)
 @ stdcall D3DX10UnsetAllDeviceObjects(ptr)
 @ stdcall D3DXBoxBoundProbe(ptr ptr ptr ptr) d3dx9_36.D3DXBoxBoundProbe
 @ stdcall D3DXColorAdjustContrast(ptr ptr float) d3dx9_36.D3DXColorAdjustContrast
