diff --git a/dlls/win32u/dib.c b/dlls/win32u/dib.c
--- a/dlls/win32u/dib.c
+++ b/dlls/win32u/dib.c
@@ -181,7 +181,7 @@
     }
     else
     {
-        WARN( "(%u): unknown/wrong size for header\n", info->biSize );
+        WARN( "(%u): unknown/wrong size for header\n", (int)info->biSize );
         return FALSE;
     }
 
@@ -493,7 +493,7 @@
     RECT rect;
 
     TRACE("%d %d %d %d <- %d %d %d %d rop %08x\n", xDst, yDst, widthDst, heightDst,
-          xSrc, ySrc, widthSrc, heightSrc, rop);
+          xSrc, ySrc, widthSrc, heightSrc, (int)rop);
 
     src_bits.ptr = (void*)bits;
     src_bits.is_copy = FALSE;
@@ -1442,7 +1442,7 @@
     height = abs( height );
 
     TRACE( "hdc=%p, init=%u, bits=%p, data=%p, coloruse=%u (bitmap: width=%d, height=%d)\n",
-           hdc, init, bits, data, coloruse, width, height );
+           hdc, (int)init, bits, data, coloruse, width, height );
 
     if (hdc == NULL)
         handle = NtGdiCreateBitmap( width, height, 1, 1, NULL );
@@ -1490,10 +1490,10 @@
     if (!(bmp = calloc( 1, sizeof(*bmp) ))) return 0;
 
     TRACE("format (%d,%d), planes %d, bpp %d, %s, size %d %s\n",
-          info->bmiHeader.biWidth, info->bmiHeader.biHeight,
+          (int)info->bmiHeader.biWidth, (int)info->bmiHeader.biHeight,
           info->bmiHeader.biPlanes, info->bmiHeader.biBitCount,
           info->bmiHeader.biCompression == BI_BITFIELDS? "BI_BITFIELDS" : "BI_RGB",
-          info->bmiHeader.biSizeImage, usage == DIB_PAL_COLORS? "PAL" : "RGB");
+          (int)info->bmiHeader.biSizeImage, usage == DIB_PAL_COLORS? "PAL" : "RGB");
 
     bmp->dib.dsBm.bmType       = 0;
     bmp->dib.dsBm.bmWidth      = info->bmiHeader.biWidth;
