diff --git a/dlls/activeds/activeds.idl b/dlls/activeds/activeds.idl
--- a/dlls/activeds/activeds.idl
+++ b/dlls/activeds/activeds.idl
@@ -19,37 +19,9 @@
 #pragma makedep register
 
 [
-    threading(both),
     helpstring("ADs LDAP Pathname Descriptor Object"),
+    progid("Pathname"),
     uuid(080d0d78-f421-11d0-a36e-00c04fb950dc),
-    progid("Pathname"),
-    version(1.0)
+    threading(both)
 ]
-coclass Pathname
-{
-    interface IADsPathname;
-}
-
-[
-    threading(both),
-    helpstring("ADs DN With String Object"),
-    uuid(334857cc-f934-11d2-ba96-00c04fb6d0d1),
-    progid("DNWithString"),
-    version(1.0)
-]
-coclass DNWithString
-{
-    interface IADsDNWithString;
-}
-
-[
-    threading(both),
-    helpstring("ADs DN With Binary Object"),
-    uuid(7e99c0a3-f935-11d2-ba96-00c04fb6d0d1),
-    progid("DNWithBinary"),
-    version(1.0)
-]
-coclass DNWithBinary
-{
-    interface IADsDNWithBinary;
-}
+coclass Pathname { interface IADsPathname; }
