diff --git a/dlls/crypt32/crypt32_private.h b/dlls/crypt32/crypt32_private.h
--- a/dlls/crypt32/crypt32_private.h
+++ b/dlls/crypt32/crypt32_private.h
@@ -185,7 +185,6 @@
 struct _context_t {
     const context_vtbl_t *vtbl;
     LONG ref;
-    BOOL deleted_from_store;
     struct WINE_CRYPTCERTSTORE *store;
     struct _context_t *linked;
     CONTEXT_PROPERTY_LIST *properties;
@@ -254,10 +253,7 @@
  DWORD dwPropID, DWORD dwFlags, const void *pvData);
 typedef BOOL (WINAPI *SerializeElementFunc)(const void *context, DWORD dwFlags,
  BYTE *pbElement, DWORD *pcbElement);
-typedef BOOL (WINAPI *DeleteContextFromStoreFunc)(const void *contex);
-typedef const void * (*FindContextByContextFunc)(HCERTSTORE store, const void *context);
-typedef const void * (WINAPI *DuplicateContextFunc)(const void *context);
-typedef void (WINAPI *FreeContextFunc)(const void *context);
+typedef BOOL (WINAPI *DeleteContextFunc)(const void *contex);
 
 /* An abstract context (certificate, CRL, or CTL) interface */
 typedef struct _WINE_CONTEXT_INTERFACE
@@ -270,10 +266,7 @@
     GetContextPropertyFunc       getProp;
     SetContextPropertyFunc       setProp;
     SerializeElementFunc         serialize;
-    DeleteContextFromStoreFunc   deleteFromStore;
-    FindContextByContextFunc     findContextByContext;
-    DuplicateContextFunc         duplicateContext;
-    FreeContextFunc              freeContext;
+    DeleteContextFunc            deleteFromStore;
 } WINE_CONTEXT_INTERFACE;
 
 extern const WINE_CONTEXT_INTERFACE *pCertInterface;
@@ -424,8 +417,6 @@
  */
 context_t *Context_CreateLinkContext(unsigned contextSize, context_t *linked, struct WINE_CRYPTCERTSTORE*);
 
-BOOL CRYPT_DeleteCertificateFromStore(PCCERT_CONTEXT pCertContext);
-
 /* Copies properties from fromContext to toContext. */
 void Context_CopyProperties(const void *to, const void *from);
 
