diff --git a/dlls/d3dx11_43/d3dx11_43_main.c b/dlls/d3dx11_43/d3dx11_43_main.c
--- a/dlls/d3dx11_43/d3dx11_43_main.c
+++ b/dlls/d3dx11_43/d3dx11_43_main.c
@@ -29,40 +29,7 @@
 #include "d3dx11core.h"
 #include "d3dx11tex.h"
 
-#include "wine/debug.h"
-
-WINE_DEFAULT_DEBUG_CHANNEL(d3dx);
-
 BOOL WINAPI D3DX11CheckVersion(UINT d3d_sdk_ver, UINT d3dx_sdk_ver)
 {
     return d3d_sdk_ver == D3D11_SDK_VERSION && d3dx_sdk_ver == D3DX11_SDK_VERSION;
 }
-
-HRESULT WINAPI D3DX11FilterTexture(ID3D11DeviceContext *context, ID3D11Resource *texture, UINT src_level, UINT filter)
-{
-    FIXME("context %p, texture %p, src_level %u, filter %#x stub!\n", context, texture, src_level, filter);
-
-    return E_NOTIMPL;
-}
-
-HRESULT WINAPI D3DX11GetImageInfoFromFileA(const char *filename, ID3DX11ThreadPump *pump, D3DX11_IMAGE_INFO *img_info,
-        HRESULT *hresult)
-{
-    FIXME("filename %s, pump %p, img_info %p, hresult %p stub!\n", debugstr_a(filename), pump, img_info, hresult);
-
-    if (!filename)
-        return E_FAIL;
-
-    return E_NOTIMPL;
-}
-
-HRESULT WINAPI D3DX11GetImageInfoFromFileW(const WCHAR *filename, ID3DX11ThreadPump *pump, D3DX11_IMAGE_INFO *img_info,
-        HRESULT *hresult)
-{
-    FIXME("filename %s, pump %p, img_info %p, hresult %p stub!\n", debugstr_w(filename), pump, img_info, hresult);
-
-    if (!filename)
-        return E_FAIL;
-
-    return E_NOTIMPL;
-}
