diff --git a/dlls/qcap/vfwcapture.c b/dlls/qcap/vfwcapture.c
--- a/dlls/qcap/vfwcapture.c
+++ b/dlls/qcap/vfwcapture.c
@@ -857,33 +857,15 @@
     return E_NOTIMPL;
 }
 
-/* Despite the name this function actually returns a list of frame intervals. */
-static HRESULT WINAPI video_control_GetFrameRateList(IAMVideoControl *iface, IPin *pin,
-        LONG index, SIZE dimensions, LONG *ret_count, LONGLONG **ret_intervals)
+static HRESULT WINAPI video_control_GetFrameRateList(IAMVideoControl *iface, IPin *pin, LONG index,
+        SIZE dimensions, LONG *list_size, LONGLONG **frame_rate)
 {
     struct vfw_capture *filter = impl_from_IAMVideoControl(iface);
-    struct get_frame_intervals_params params;
-    unsigned int count;
-
-    TRACE("filter %p, pin %p, index %ld, dimensions (%ldx%ld), ret_count %p, ret_intervals %p.\n",
-            filter, pin, index, dimensions.cx, dimensions.cy, ret_count, ret_intervals);
-
-    params.device = filter->device;
-    params.caps_index = index;
-    params.width = dimensions.cx;
-    params.height = dimensions.cy;
-    params.intervals = NULL;
-    params.count = &count;
-    V4L_CALL( get_frame_intervals, &params );
-
-    if (!(params.intervals = CoTaskMemAlloc(count * sizeof(LONGLONG))))
-        return E_OUTOFMEMORY;
-
-    V4L_CALL( get_frame_intervals, &params );
-
-    *ret_intervals = params.intervals;
-    *ret_count = count;
-    return S_OK;
+
+    FIXME("filter %p, pin %p, index %ld, dimensions (%ldx%ld), list size %p, frame rate %p, stub.\n",
+            filter, pin, index, dimensions.cx, dimensions.cy, list_size, frame_rate);
+
+    return E_NOTIMPL;
 }
 
 static const IAMVideoControlVtbl IAMVideoControl_VTable =
