diff --git a/dlls/mfplat/buffer.c b/dlls/mfplat/buffer.c
--- a/dlls/mfplat/buffer.c
+++ b/dlls/mfplat/buffer.c
@@ -1004,6 +1004,7 @@
     {
         ID3D11DeviceContext_CopySubresourceRegion(immediate_context, (ID3D11Resource *)buffer->dxgi_surface.texture,
                 buffer->dxgi_surface.sub_resource_idx, 0, 0, 0, (ID3D11Resource *)buffer->dxgi_surface.rb_texture, 0, NULL);
+        ID3D11DeviceContext_Flush(immediate_context);
     }
 
     ID3D11DeviceContext_Release(immediate_context);
@@ -1352,7 +1353,6 @@
 
     if (!(buffer->data = _aligned_malloc(max_length, alignment)))
         return E_OUTOFMEMORY;
-    memset(buffer->data, 0, max_length);
 
     buffer->IMFMediaBuffer_iface.lpVtbl = vtbl;
     buffer->refcount = 1;
@@ -1405,7 +1405,6 @@
             return copy_image_imc2;
 
         case MAKEFOURCC('N','V','1','2'):
-        case MAKEFOURCC('P','0','1','0'):
             return copy_image_nv12;
 
         default:
@@ -1455,9 +1454,6 @@
         case MAKEFOURCC('N','V','1','2'):
             plane_size = stride * height * 3 / 2;
             break;
-        case MAKEFOURCC('P','0','1','0'):
-            plane_size = width * height * 3;
-            break;
         default:
             plane_size = stride * height;
     }
@@ -1496,7 +1492,6 @@
         case MAKEFOURCC('N','V','1','1'):
         case MAKEFOURCC('I','4','2','0'):
         case MAKEFOURCC('I','Y','U','V'):
-        case MAKEFOURCC('P','0','1','0'):
             max_length = pitch * height * 3 / 2;
             break;
         default:
