diff --git a/dlls/oleacc/client.c b/dlls/oleacc/client.c
--- a/dlls/oleacc/client.c
+++ b/dlls/oleacc/client.c
@@ -23,6 +23,7 @@
 #include "commctrl.h"
 
 #include "wine/debug.h"
+#include "wine/heap.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(oleacc);
 
@@ -143,7 +144,7 @@
     TRACE("(%p) ref = %lu\n", This, ref);
 
     if(!ref)
-        free(This);
+        heap_free(This);
     return ref;
 }
 
@@ -840,13 +841,13 @@
         return E_ACCESSDENIED;
 
     len = SendMessageW(client->hwnd, WM_GETTEXTLENGTH, 0, 0);
-    buf = calloc(len + 1, sizeof(*buf));
+    buf = heap_alloc_zero((len + 1) * sizeof(*buf));
     if(!buf)
         return E_OUTOFMEMORY;
 
     SendMessageW(client->hwnd, WM_GETTEXT, len + 1, (LPARAM)buf);
     *value_out = SysAllocString(buf);
-    free(buf);
+    heap_free(buf);
     return S_OK;
 }
 
@@ -909,7 +910,7 @@
     if(!IsWindow(hwnd))
         return E_FAIL;
 
-    client = calloc(1, sizeof(Client));
+    client = heap_alloc_zero(sizeof(Client));
     if(!client)
         return E_OUTOFMEMORY;
 
