diff --git a/dlls/secur32/schannel.c b/dlls/secur32/schannel.c
--- a/dlls/secur32/schannel.c
+++ b/dlls/secur32/schannel.c
@@ -65,6 +65,7 @@
     enum control_token control_token;
     unsigned int alert_type;
     unsigned int alert_number;
+    BOOL rehandshake_requested;
 };
 
 static struct schan_handle *schan_handle_table;
@@ -898,7 +899,7 @@
             buffer = &pInput->pBuffers[idx];
             ptr = buffer->pvBuffer;
 
-            if (buffer->cbBuffer < ctx->header_size)
+            if (buffer->cbBuffer < ctx->header_size && !ctx->rehandshake_requested)
             {
                 TRACE("Expected at least %Iu bytes, but buffer only contains %lu bytes.\n",
                       ctx->header_size, buffer->cbBuffer);
@@ -915,7 +916,7 @@
                 ptr += record_size;
             }
 
-            if (!expected_size)
+            if (!expected_size && !ctx->rehandshake_requested)
             {
                 TRACE("Expected at least %Iu bytes, but buffer only contains %lu bytes.\n",
                       max(ctx->header_size, record_size), buffer->cbBuffer);
@@ -967,6 +968,7 @@
     params.alert_type = ctx->alert_type;
     params.alert_number = ctx->alert_number;
     ctx->control_token = CONTROL_TOKEN_NONE;
+    ctx->rehandshake_requested = FALSE;
     ret = GNUTLS_CALL( handshake, &params );
 
     if (output_buffer_idx != -1)
@@ -1569,6 +1571,7 @@
     buffer->BufferType = SECBUFFER_STREAM_HEADER;
     buffer->cbBuffer = ctx->header_size;
 
+    if (status == SEC_I_RENEGOTIATE) ctx->rehandshake_requested = TRUE;
     return status;
 }
 
