diff --git a/dlls/adsldp/adsldp.idl b/dlls/adsldp/adsldp.idl
--- a/dlls/adsldp/adsldp.idl
+++ b/dlls/adsldp/adsldp.idl
@@ -21,35 +21,24 @@
 #pragma makedep register
 
 [
-    threading(both),
     helpstring("LDAP Provider Object"),
     progid("LDAP"),
-    uuid(228d9a81-c302-11cf-9aa4-00aa004a5691)
+    uuid(228d9a81-c302-11cf-9aa4-00aa004a5691),
+    threading(both)
 ]
-coclass LDAP
-{
-    interface IParseDisplayName;
-}
+coclass LDAP { interface IParseDisplayName; }
 
 [
-    threading(both),
     helpstring("LDAP Namespace Object"),
     progid("LDAPNamespace"),
     uuid(228d9a82-c302-11cf-9aa4-00aa004a5691),
+    threading(both)
 ]
-coclass ADs
-{
-    interface IADs;
-}
+coclass ADs { }
 
 [
-    threading(both),
     helpstring("AD System Info Object"),
     uuid(50b6327f-afd1-11d2-9cb9-0000f87a369e),
-    progid("ADSystemInfo"),
-    version(1.0)
+    threading(both)
 ]
-coclass ADSystemInfo
-{
-    interface IADsADSystemInfo;
-}
+coclass ADSystemInfo { }
