diff --git a/dlls/mscoree/corruntimehost.c b/dlls/mscoree/corruntimehost.c
--- a/dlls/mscoree/corruntimehost.c
+++ b/dlls/mscoree/corruntimehost.c
@@ -832,7 +832,7 @@
     IHostControl *pHostControl)
 {
     FIXME("(%p,%p)\n", iface, pHostControl);
-    return S_OK;
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI CLRRuntimeHost_GetCLRControl(ICLRRuntimeHost* iface,
@@ -1103,7 +1103,7 @@
 
 # define CAN_FIXUP_VTABLE 1
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 
 struct vtable_fixup_thunk
 {
@@ -1142,13 +1142,13 @@
     NULL
 };
 
-#pragma pack(pop)
+#include "poppack.h"
 
 #elif __x86_64__ /* !__i386__ */
 
 # define CAN_FIXUP_VTABLE 1
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 
 struct vtable_fixup_thunk
 {
@@ -1210,7 +1210,7 @@
     {0x48,0x8B,0x00,0xFF,0xE0}
 };
 
-#pragma pack(pop)
+#include "poppack.h"
 
 #else /* !__i386__ && !__x86_64__ */
 
