diff --git a/dlls/browseui/browseui_main.c b/dlls/browseui/browseui_main.c
--- a/dlls/browseui/browseui_main.c
+++ b/dlls/browseui/browseui_main.c
@@ -28,6 +28,7 @@
 #include "windef.h"
 #include "winbase.h"
 #include "winreg.h"
+#include "shlwapi.h"
 #include "shlguid.h"
 #include "rpcproxy.h"
 
@@ -180,6 +181,35 @@
 }
 
 /***********************************************************************
+ *              DllGetVersion (BROWSEUI.@)
+ */
+HRESULT WINAPI DllGetVersion(DLLVERSIONINFO *info)
+{
+    if(info->cbSize == sizeof(DLLVERSIONINFO) ||
+       info->cbSize == sizeof(DLLVERSIONINFO2))
+    {
+        /* this is what IE6 on Windows 98 reports */
+        info->dwMajorVersion = 6;
+        info->dwMinorVersion = 0;
+        info->dwBuildNumber = 2600;
+        info->dwPlatformID = DLLVER_PLATFORM_WINDOWS;
+        if(info->cbSize == sizeof(DLLVERSIONINFO2))
+        {
+            DLLVERSIONINFO2 *info2 = (DLLVERSIONINFO2*) info;
+            info2->dwFlags = 0;
+            info2->ullVersion = MAKEDLLVERULL(info->dwMajorVersion,
+                                              info->dwMinorVersion,
+                                              info->dwBuildNumber,
+                                              0); /* FIXME: correct hotfix number */
+        }
+        return S_OK;
+    }
+
+    WARN("wrong DLLVERSIONINFO size from app.\n");
+    return E_INVALIDARG;
+}
+
+/***********************************************************************
  *              DllGetClassObject (BROWSEUI.@)
  */
 HRESULT WINAPI DllGetClassObject(REFCLSID clsid, REFIID iid, LPVOID *ppvOut)
