diff --git a/dlls/httpapi/tests/httpapi.c b/dlls/httpapi/tests/httpapi.c
--- a/dlls/httpapi/tests/httpapi.c
+++ b/dlls/httpapi/tests/httpapi.c
@@ -1354,42 +1354,7 @@
         }
         ok(ret == ERROR_SHARING_VIOLATION, "Failed to add %s, error %u.\n", debugstr_w(url), ret);
     }
-    ok(!ret, "Got error %u.\n", ret);
-
-    s = create_client_socket(port);
-    sprintf(req_text, simple_req, port);
-    ret = send(s, req_text, strlen(req_text), 0);
-    ok(ret == strlen(req_text), "send() returned %d.\n", ret);
-
-    memset(req_buffer, 0xcc, sizeof(req_buffer));
-    ret = HttpReceiveHttpRequest(queue, HTTP_NULL_ID, 0, (HTTP_REQUEST *)req, sizeof(req_buffer), &ret_size, NULL);
-    ok(!ret, "Got error %u.\n", ret);
-    ok(ret_size > sizeof(*req), "Got size %lu.\n", ret_size);
-
-    send_response_v1(queue, req->RequestId, s);
-
-    ret = HttpRemoveUrl(queue, url);
-    ok(!ret, "Got error %u.\n", ret);
-    closesocket(s);
-    ret = CloseHandle(queue);
-    ok(ret, "Failed to close queue handle, error %lu.\n", GetLastError());
-
-    ret = HttpCreateHttpHandle(&queue, 0);
-    ok(!ret, "Got error %u.\n", ret);
-
-    for (port = 51000; port < 52000; ++port)
-    {
-        swprintf(url, ARRAY_SIZE(url), L"http://*:%u/", port);
-        if (!(ret = HttpAddUrl(queue, url, NULL)))
-            break;
-        if (ret == ERROR_ACCESS_DENIED)
-        {
-            skip("Not enough permissions to bind to all URLs.\n");
-            CloseHandle(queue);
-            return;
-        }
-        ok(ret == ERROR_SHARING_VIOLATION, "Failed to add %s, error %u.\n", debugstr_w(url), ret);
-    }
+
     ok(!ret, "Got error %u.\n", ret);
 
     s = create_client_socket(port);
