diff --git a/dlls/imm32/imm.c b/dlls/imm32/imm.c
--- a/dlls/imm32/imm.c
+++ b/dlls/imm32/imm.c
@@ -3073,18 +3073,13 @@
     while (ctx->dwNumMsgBuf--)
     {
         TRANSMSG *msgs, msg;
-        if (!(msgs = ImmLockIMCC( ctx->hMsgBuf )))
-        {
-            ImmUnlockIMC( himc );
-            return FALSE;
-        }
+        if (!(msgs = ImmLockIMCC( ctx->hMsgBuf ))) return FALSE;
         msg = msgs[0];
         memmove( msgs, msgs + 1, ctx->dwNumMsgBuf * sizeof(*msgs) );
         ImmUnlockIMCC( ctx->hMsgBuf );
         SendMessageW( ctx->hWnd, msg.message, msg.wParam, msg.lParam );
     }
     ctx->dwNumMsgBuf++;
-    ImmUnlockIMC( himc );
 
     return TRUE;
 }
@@ -3347,23 +3342,3 @@
     SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
     return FALSE;
 }
-
-/***********************************************************************
- *      CtfImmHideToolbarWnd (IMM32.@)
- */
-DWORD WINAPI CtfImmHideToolbarWnd(void)
-{
-    FIXME("(): stub\n");
-    SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
-    return 0;
-}
-
-/***********************************************************************
- *      CtfImmRestoreToolbarWnd (IMM32.@)
- */
-DWORD WINAPI CtfImmRestoreToolbarWnd(DWORD unknown)
-{
-    FIXME("%lx: stub\n", unknown);
-    SetLastError(ERROR_CALL_NOT_IMPLEMENTED);
-    return 0;
-}
