diff --git a/dlls/comdlg32/fontdlg.c b/dlls/comdlg32/fontdlg.c
--- a/dlls/comdlg32/fontdlg.c
+++ b/dlls/comdlg32/fontdlg.c
@@ -32,6 +32,7 @@
 #include "commdlg.h"
 #include "dlgs.h"
 #include "wine/debug.h"
+#include "wine/heap.h"
 #include "cderr.h"
 #include "cdlg.h"
 
@@ -1141,11 +1142,11 @@
     if((lpcfw->Flags & CF_USESTYLE) && lpcfw->lpszStyle) {
         len = WideCharToMultiByte(CP_ACP, 0, lpcfw->lpszStyle, -1, NULL, 0, 0, 0);
         WideCharToMultiByte(CP_ACP, 0, lpcfw->lpszStyle, -1, lpcfa->lpszStyle, len, 0, 0);
-        free(lpcfw->lpszStyle);
-    }
-
-    free(lpcfw->lpLogFont);
-    free(lpcfw);
+        heap_free(lpcfw->lpszStyle);
+    }
+
+    heap_free(lpcfw->lpLogFont);
+    heap_free(lpcfw);
     SetPropW(hwnd, strWineFontData, 0);
 
     return TRUE;
@@ -1208,16 +1209,16 @@
         lpcfa=(LPCHOOSEFONTA)lParam;
         SetPropW(hDlg, strWineFontData_a, (HANDLE)lParam);
 
-        lpcfw = malloc(sizeof(*lpcfw));
+        lpcfw = heap_alloc(sizeof(*lpcfw));
         memcpy(lpcfw, lpcfa, sizeof(CHOOSEFONTA));
-        lpcfw->lpLogFont = malloc(sizeof(*lpcfw->lpLogFont));
+        lpcfw->lpLogFont = heap_alloc(sizeof(*lpcfw->lpLogFont));
         memcpy(lpcfw->lpLogFont, lpcfa->lpLogFont, sizeof(LOGFONTA));
         MultiByteToWideChar(CP_ACP, 0, lpcfa->lpLogFont->lfFaceName,
                             LF_FACESIZE, lpcfw->lpLogFont->lfFaceName, LF_FACESIZE);
 
         if((lpcfa->Flags & CF_USESTYLE) && lpcfa->lpszStyle)  {
             len = MultiByteToWideChar(CP_ACP, 0, lpcfa->lpszStyle, -1, NULL, 0);
-            lpcfw->lpszStyle = malloc(len * sizeof(WCHAR));
+            lpcfw->lpszStyle = heap_alloc(len * sizeof(WCHAR));
             MultiByteToWideChar(CP_ACP, 0, lpcfa->lpszStyle, -1, lpcfw->lpszStyle, len);
         }
 
