diff --git a/dlls/user32/combo.c b/dlls/user32/combo.c
--- a/dlls/user32/combo.c
+++ b/dlls/user32/combo.c
@@ -230,10 +230,10 @@
     measureItem.CtlID      = id;
     measureItem.itemID     = -1;
     measureItem.itemWidth  = clientRect.right;
-    measureItem.itemHeight = item_height - 2; /* ownerdrawn cb is taller */
+    measureItem.itemHeight = item_height - 6; /* ownerdrawn cb is taller */
     measureItem.itemData   = 0;
     SendMessageW(lphc->owner, WM_MEASUREITEM, id, (LPARAM)&measureItem);
-    item_height = 2 + measureItem.itemHeight;
+    item_height = 6 + measureItem.itemHeight;
 
     /*
      * Send a second one in the case of a fixed ownerdraw list to calculate the
@@ -982,7 +982,7 @@
    if( !(lphc->wState & CBF_NOREDRAW) )
      NtUserRedrawWindow( lphc->self, NULL, 0, RDW_INVALIDATE | RDW_ERASE | RDW_UPDATENOW );
 
-   NtUserEnableWindow( lphc->hWndLBox, TRUE );
+   EnableWindow( lphc->hWndLBox, TRUE );
    if (GetCapture() != lphc->self)
       NtUserSetCapture(lphc->hWndLBox);
 }
@@ -1817,8 +1817,8 @@
 		return COMBO_ItemOp(lphc, message, lParam);
 	case WM_ENABLE:
 		if( lphc->wState & CBF_EDIT )
-		    NtUserEnableWindow( lphc->hWndEdit, (BOOL)wParam );
-		NtUserEnableWindow( lphc->hWndLBox, (BOOL)wParam );
+		    EnableWindow( lphc->hWndEdit, (BOOL)wParam );
+		EnableWindow( lphc->hWndLBox, (BOOL)wParam );
 
 		/* Force the control to repaint when the enabled state changes. */
 		NtUserInvalidateRect(lphc->self, NULL, TRUE);
@@ -1896,6 +1896,11 @@
         case WM_CTLCOLORSTATIC:
             if (lphc->owner)
                 return SendMessageW(lphc->owner, message, wParam, lParam);
+            break;
+
+        case WM_GETOBJECT:
+            if ((LONG)lParam == OBJID_QUERYCLASSNAMEIDX)
+                return 0x10005;
             break;
 
 	/* Combo messages */
