diff --git a/dlls/wow64/struct32.h b/dlls/wow64/struct32.h
--- a/dlls/wow64/struct32.h
+++ b/dlls/wow64/struct32.h
@@ -75,13 +75,6 @@
 
 typedef struct
 {
-    WORD  Level;
-    WORD  Sbz;
-    ULONG ObjectType;
-} OBJECT_TYPE_LIST32;
-
-typedef struct
-{
     CONTEXT_CHUNK All;
     CONTEXT_CHUNK Legacy;
     CONTEXT_CHUNK XState;
@@ -332,6 +325,30 @@
     ULONG                            TimeDateStamp;
     ULONG                            DefaultBase;
 } RTL_PROCESS_MODULE_INFORMATION_EX32;
+
+typedef struct
+{
+    ULONG Flags;
+    union
+    {
+        ULONG TlsVector;
+        ULONG TlsModulePointer;
+    };
+    ULONG ThreadId;
+} THREAD_TLS_INFORMATION32;
+
+typedef struct
+{
+    ULONG Flags;
+    ULONG OperationType;
+    ULONG ThreadDataCount;
+    union
+    {
+        ULONG TlsIndex;
+        ULONG TlsVectorLength;
+    };
+    THREAD_TLS_INFORMATION32 ThreadData[1];
+} PROCESS_TLS_INFORMATION32;
 
 typedef struct
 {
