diff --git a/dlls/bthprops.cpl/bthprops.cpl.spec b/dlls/bthprops.cpl/bthprops.cpl.spec
--- a/dlls/bthprops.cpl/bthprops.cpl.spec
+++ b/dlls/bthprops.cpl/bthprops.cpl.spec
@@ -1,12 +1,12 @@
 @ stub BluetoothAddressToString
 @ stub BluetoothAuthenticateDevice
-@ stdcall -import BluetoothAuthenticateDeviceEx(ptr ptr ptr ptr long)
+@ stub BluetoothAuthenticateDeviceEx
 @ stub BluetoothAuthenticateMultipleDevices
 @ stub BluetoothAuthenticationAgent
 @ stub BluetoothDisconnectDevice
 @ stub BluetoothDisplayDeviceProperties
-@ stdcall -import BluetoothEnableDiscovery(ptr long)
-@ stdcall -import BluetoothEnableIncomingConnections(ptr long)
+@ stub BluetoothEnableDiscovery
+@ stub BluetoothEnableIncomingConnections
 @ stub BluetoothEnumerateInstalledServices
 @ stub BluetoothEnumerateInstalledServicesEx
 @ stub BluetoothFindBrowseGroupClose
@@ -34,16 +34,16 @@
 @ stub BluetoothFindProtocolEntryClose
 @ stdcall -import BluetoothFindRadioClose(ptr)
 @ stub BluetoothFindServiceClose
-@ stdcall BluetoothGetDeviceInfo(ptr long)
+@ stub BluetoothGetDeviceInfo
 @ stdcall -import BluetoothGetRadioInfo(ptr ptr)
-@ stdcall -import BluetoothIsConnectable(ptr)
-@ stdcall -import BluetoothIsDiscoverable(ptr)
+@ stub BluetoothIsConnectable
+@ stub BluetoothIsDiscoverable
 @ stub BluetoothIsVersionAvailable
 @ stub BluetoothMapClassOfDeviceToImageIndex
 @ stub BluetoothMapClassOfDeviceToString
 @ stub BluetoothRegisterForAuthentication
 @ stdcall -import BluetoothRegisterForAuthenticationEx(ptr ptr ptr ptr)
-@ stdcall -import BluetoothRemoveDevice(ptr)
+@ stub BluetoothRemoveDevice
 @ stdcall -import BluetoothSdpEnumAttributes(ptr long ptr ptr)
 @ stdcall -import BluetoothSdpGetAttributeValue(ptr long long ptr)
 @ stdcall -import BluetoothSdpGetContainerElementData(ptr long ptr ptr)
@@ -52,7 +52,7 @@
 @ stub BluetoothSelectDevices
 @ stub BluetoothSelectDevicesFree
 @ stub BluetoothSendAuthenticationResponse
-@ stdcall -import BluetoothSendAuthenticationResponseEx(ptr ptr)
+@ stub BluetoothSendAuthenticationResponseEx
 @ stub BluetoothSetLocalServiceInfo
 @ stub BluetoothSetServiceState
 @ stdcall -import BluetoothUnregisterAuthentication(long)
