diff --git a/dlls/win32u/bitblt.c b/dlls/win32u/bitblt.c
--- a/dlls/win32u/bitblt.c
+++ b/dlls/win32u/bitblt.c
@@ -557,7 +557,7 @@
 
         TRACE("dst %p log=%d,%d %dx%d phys=%d,%d %dx%d vis=%s  rop=%06x\n",
               hdc, dst.log_x, dst.log_y, dst.log_width, dst.log_height,
-              dst.x, dst.y, dst.width, dst.height, wine_dbgstr_rect(&dst.visrect), rop );
+              dst.x, dst.y, dst.width, dst.height, wine_dbgstr_rect(&dst.visrect), (int)rop );
 
         if (!ret)
         {
@@ -624,7 +624,7 @@
               hdcSrc, src.log_x, src.log_y, src.log_width, src.log_height,
               src.x, src.y, src.width, src.height, wine_dbgstr_rect(&src.visrect),
               hdcDst, dst.log_x, dst.log_y, dst.log_width, dst.log_height,
-              dst.x, dst.y, dst.width, dst.height, wine_dbgstr_rect(&dst.visrect), rop );
+              dst.x, dst.y, dst.width, dst.height, wine_dbgstr_rect(&dst.visrect), (int)rop );
 
         if (!ret)
         {
@@ -857,7 +857,6 @@
     int oldStretchMode;
     DIBSECTION dib;
     DC *dc_src;
-    DC *dc_work;
 
     if(widthDest < 0 || heightDest < 0 || widthSrc < 0 || heightSrc < 0) {
         TRACE("Cannot mirror\n");
@@ -891,11 +890,6 @@
     }
     else bmpWork = NtGdiCreateCompatibleBitmap( hdcDest, widthDest, heightDest );
     oldWork = NtGdiSelectBitmap(hdcWork, bmpWork);
-
-    if (!(dc_work = get_dc_ptr(hdcWork))) goto error;
-    dc_work->attr->stretch_blt_mode = COLORONCOLOR;
-    release_dc_ptr( dc_work );
-
     if (!NtGdiStretchBlt( hdcWork, 0, 0, widthDest, heightDest, hdcSrc, xSrc, ySrc,
                           widthSrc, heightSrc, SRCCOPY, 0 ))
     {
@@ -1074,7 +1068,7 @@
 
     TRACE("hdcSrc=%p %d,%d,%dx%d -> hdcDest=%p %d,%d,%d,%d,%d,%d\n",
           hdcSrc, nXSrc, nYSrc, nWidth, nHeight, hdcDest,
-          plg[0].x, plg[0].y, plg[1].x, plg[1].y, plg[2].x, plg[2].y);
+          (int)plg[0].x, (int)plg[0].y, (int)plg[1].x, (int)plg[1].y, (int)plg[2].x, (int)plg[2].y);
 
     /* X components */
     xf.eM11 = (plg[1].x*(rect[2].y - rect[0].y) - plg[2].x*(rect[1].y - rect[0].y) - plg[0].x*(rect[2].y - rect[1].y)) / det;
