diff --git a/dlls/windows.gaming.input/provider.idl b/dlls/windows.gaming.input/provider.idl
--- a/dlls/windows.gaming.input/provider.idl
+++ b/dlls/windows.gaming.input/provider.idl
@@ -24,6 +24,7 @@
 
 import "propidl.idl";
 import "inspectable.idl";
+import "asyncinfo.idl";
 import "eventtoken.idl";
 import "windowscontracts.idl";
 import "windows.foundation.idl";
@@ -45,6 +46,9 @@
     interface IWineGameControllerProvider;
     runtimeclass WineGameControllerProvider;
 
+    /* type-pruning version of AsyncOperationCompletedHandler<T> */
+    delegate HRESULT WineAsyncOperationCompletedHandler([in] IInspectable *async, [in] AsyncStatus status);
+
     enum WineGameControllerType
     {
         Joystick = 0,
@@ -190,12 +194,22 @@
     interface IWineForceFeedbackEffectImpl : IUnknown
         requires Windows.Gaming.Input.ForceFeedback.IForceFeedbackEffect
     {
-        HRESULT SetParameters([in] WineForceFeedbackEffectParameters parameters,
-                              [in, optional] WineForceFeedbackEffectEnvelope *envelope);
-    }
-
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
+        [propput] HRESULT Parameters([in] WineForceFeedbackEffectParameters parameters,
+                                     [in, optional] WineForceFeedbackEffectEnvelope *envelope);
+    }
+
+    [
+        uuid(83f377ee-c799-11ec-9d64-0242ac120002)
+    ]
+    interface IWineAsyncInfoImpl : IUnknown
+    {
+        [propput] HRESULT Completed([in] WineAsyncOperationCompletedHandler *handler);
+        [propget] HRESULT Completed([out, retval] WineAsyncOperationCompletedHandler **handler);
+        [propget] HRESULT Result([out, retval] PROPVARIANT *result);
+        HRESULT Start();
+    }
+
+    [
         marshaling_behavior(agile),
         threading(both)
     ]
