diff --git a/dlls/mshtml/htmllink.c b/dlls/mshtml/htmllink.c
--- a/dlls/mshtml/htmllink.c
+++ b/dlls/mshtml/htmllink.c
@@ -24,7 +24,6 @@
 #include "winbase.h"
 #include "winuser.h"
 #include "ole2.h"
-#include "mshtmdid.h"
 
 #include "wine/debug.h"
 
@@ -371,18 +370,6 @@
     .get_disabled          = HTMLLinkElementImpl_get_disabled,
 };
 
-static void HTMLLinkElement_init_dispex_info(dispex_data_t *info, compat_mode_t mode)
-{
-    static const dispex_hook_t link_hooks[] = {
-        {DISPID_IHTMLLINKELEMENT_STYLESHEET, .noattr = TRUE},
-        {DISPID_IHTMLLINKELEMENT_READYSTATE, .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    dispex_info_add_interface(info, IHTMLLinkElement_tid, link_hooks);
-
-    HTMLElement_init_dispex_info(info, mode);
-}
-
 static const event_target_vtbl_t HTMLLinkElement_event_target_vtbl = {
     {
         HTMLELEMENT_DISPEX_VTBL_ENTRIES,
@@ -395,12 +382,17 @@
     .handle_event       = HTMLElement_handle_event
 };
 
+static const tid_t HTMLLinkElement_iface_tids[] = {
+    IHTMLLinkElement_tid,
+    0
+};
 dispex_static_data_t HTMLLinkElement_dispex = {
-    .id           = OBJID_HTMLLinkElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLLinkElement,
+    .prototype_id = PROT_HTMLElement,
     .vtbl         = &HTMLLinkElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLLinkElement_tid,
-    .init_info    = HTMLLinkElement_init_dispex_info,
+    .iface_tids   = HTMLLinkElement_iface_tids,
+    .init_info    = HTMLElement_init_dispex_info,
 };
 
 HRESULT HTMLLinkElement_Create(HTMLDocumentNode *doc, nsIDOMElement *nselem, HTMLElement **elem)
