diff --git a/dlls/win32u/mapping.c b/dlls/win32u/mapping.c
--- a/dlls/win32u/mapping.c
+++ b/dlls/win32u/mapping.c
@@ -190,6 +190,7 @@
         {
             double x = points->x;
             double y = points->y;
+            *points = map_dpi_point( *points, dc->dpi_to, dc->dpi_from );
             points->x = GDI_ROUND( x * dc->xformVport2World.eM11 +
                                    y * dc->xformVport2World.eM21 +
                                    dc->xformVport2World.eDx );
@@ -247,6 +248,22 @@
         ret = TRUE;
         break;
 
+    case NtGdiLPtoDPRaw:
+        for (i = 0; i < count; i++)
+        {
+            double x = points_in[i].x;
+            double y = points_in[i].y;
+            points_out[i].x = GDI_ROUND( x * dc->xformWorld2Vport.eM11 +
+                                         y * dc->xformWorld2Vport.eM21 +
+                                         dc->xformWorld2Vport.eDx );
+            points_out[i].y = GDI_ROUND( x * dc->xformWorld2Vport.eM12 +
+                                         y * dc->xformWorld2Vport.eM22 +
+                                         dc->xformWorld2Vport.eDy );
+            points_out[i] = map_dpi_point( points_out[i], dc->dpi_from, dc->dpi_to );
+        }
+        ret = TRUE;
+        break;
+
     default:
         WARN( "invalid mode %x\n", mode );
         break;
@@ -274,6 +291,7 @@
         points->y = GDI_ROUND( x * dc->xformWorld2Vport.eM12 +
                                y * dc->xformWorld2Vport.eM22 +
                                dc->xformWorld2Vport.eDy );
+        *points = map_dpi_point( *points, dc->dpi_from, dc->dpi_to );
         points++;
     }
 }
@@ -433,7 +451,7 @@
                                        DWORD horz_size, DWORD vert_size )
 {
     DC * dc;
-    TRACE("(%p %d %d %d %d)\n", hdc, horz_res, vert_res, horz_size, vert_size);
+    TRACE("(%p %d %d %d %d)\n", hdc, (int)horz_res, (int)vert_res, (int)horz_size, (int)vert_size);
 
     if (!horz_res || !vert_res || !horz_size || !vert_size)
     {
