diff --git a/dlls/mfplat/mediatype.c b/dlls/mfplat/mediatype.c
--- a/dlls/mfplat/mediatype.c
+++ b/dlls/mfplat/mediatype.c
@@ -2728,7 +2728,6 @@
     { &MFVideoFormat_IYUV,          12, 0, 0, 1, -1 },
     { &MFVideoFormat_NV11,          12, 0, 0, 1, -1 },
     { &MFVideoFormat_NV12,          12, 0, 0, 1, -1 },
-    { &MFVideoFormat_P010,          24, 0, 0, 1, -1 },
     { &MFVideoFormat_D16,           16, 3, 0, 0, -1 },
     { &MFVideoFormat_L16,           16, 3, 0, 0, -1 },
     { &MFVideoFormat_UYVY,          16, 0, 0, 1, -1 },
@@ -2763,7 +2762,6 @@
 static unsigned int mf_get_stride_for_format(const struct uncompressed_video_format *format, unsigned int width)
 {
     if (format->bpp < 8) return (width * format->bpp) / 8;
-    else if (IsEqualGUID(format->subtype, &MFVideoFormat_P010)) return width * 2;
     return (width * (format->bpp / 8) + format->alignment) & ~format->alignment;
 }
 
@@ -2832,10 +2830,6 @@
         case MAKEFOURCC('I','Y','U','V'):
             /* 2 x 2 block, interleaving UV for half the height */
             *size = ((width + 1) & ~1) * height * 3 / 2;
-            break;
-        case MAKEFOURCC('P','0','1','0'):
-            /* width is rounded up to a multiple of two */
-            *size = ((width + 1) & ~1) * height * 3;
             break;
         case MAKEFOURCC('N','V','1','1'):
             *size = ((width + 3) & ~3) * height * 3 / 2;
@@ -2881,7 +2875,6 @@
         case MAKEFOURCC('I','4','2','0'):
         case MAKEFOURCC('I','Y','U','V'):
         case MAKEFOURCC('N','V','1','1'):
-        case MAKEFOURCC('P','0','1','0'):
             *size = stride * height * 3 / 2;
             break;
         default:
@@ -4418,13 +4411,6 @@
         else if (IsEqualGUID(&am_type->formattype, &FORMAT_WaveFormatEx)
                 && am_type->cbFormat >= sizeof(WAVEFORMATEX))
             hr = MFInitMediaTypeFromWaveFormatEx(media_type, (WAVEFORMATEX *)am_type->pbFormat, am_type->cbFormat);
-        else if (IsEqualGUID(&am_type->formattype, &GUID_NULL))
-        {
-            mediatype_set_guid(media_type, &MF_MT_MAJOR_TYPE, &MFMediaType_Audio, &hr);
-            mediatype_set_guid(media_type, &MF_MT_SUBTYPE, &am_type->subtype, &hr);
-            mediatype_set_guid(media_type, &MF_MT_AM_FORMAT_TYPE, &GUID_NULL, &hr);
-            mediatype_set_uint32(media_type, &MF_MT_ALL_SAMPLES_INDEPENDENT, 1, &hr);
-        }
         else
         {
             FIXME("Unsupported format type %s / size %ld.\n", debugstr_guid(&am_type->formattype), am_type->cbFormat);
