diff --git a/dlls/wpcap/unixlib.h b/dlls/wpcap/unixlib.h
--- a/dlls/wpcap/unixlib.h
+++ b/dlls/wpcap/unixlib.h
@@ -133,7 +133,7 @@
 
 struct dump_params
 {
-    UINT64 handle;
+    unsigned char *user;
     const struct pcap_pkthdr_win32 *hdr;
     const unsigned char *packet;
 };
@@ -222,9 +222,7 @@
 {
     UINT64 handle;
     struct pcap_pkthdr_win32 *hdr;
-    const unsigned char *data;
-    unsigned char *buf;
-    UINT32 bufsize;
+    const unsigned char **data;
 };
 
 struct open_live_params
@@ -262,12 +260,6 @@
     int enable;
 };
 
-struct set_immediate_mode_params
-{
-    UINT64 handle;
-    int mode;
-};
-
 struct set_rfmon_params
 {
     UINT64 handle;
@@ -344,7 +336,6 @@
 
 enum pcap_funcs
 {
-    unix_process_attach,
     unix_activate,
     unix_breakloop,
     unix_bufsize,
@@ -375,7 +366,6 @@
     unix_sendpacket,
     unix_set_buffer_size,
     unix_set_datalink,
-    unix_set_immediate_mode,
     unix_set_promisc,
     unix_set_rfmon,
     unix_set_snaplen,
