diff --git a/dlls/comctl32/Makefile.in b/dlls/comctl32/Makefile.in
--- a/dlls/comctl32/Makefile.in
+++ b/dlls/comctl32/Makefile.in
@@ -1,11 +1,13 @@
-EXTRADEFS = -D_COMCTL32_ -D__WINE_COMCTL32_VERSION=5
+EXTRADEFS = -D_COMCTL32_
 MODULE    = comctl32.dll
 IMPORTLIB = comctl32
-IMPORTS   = uuid user32 gdi32 advapi32 imm32 kernelbase oleaut32
-DELAYIMPORTS = oleacc winmm
+IMPORTS   = uuid user32 gdi32 advapi32 imm32 kernelbase oleacc oleaut32
+DELAYIMPORTS = winmm uxtheme
 
 SOURCES = \
 	animate.c \
+	button.c \
+	combo.c \
 	comboex.c \
 	comctl32.rc \
 	commctrl.c \
@@ -13,6 +15,7 @@
 	dpa.c \
 	draglist.c \
 	dsa.c \
+	edit.c \
 	flatsb.c \
 	header.c \
 	hotkey.c \
@@ -28,6 +31,7 @@
 	idc_movebutton.svg \
 	imagelist.c \
 	ipaddress.c \
+	listbox.c \
 	listview.c \
 	monthcal.c \
 	nativefont.c \
@@ -36,12 +40,14 @@
 	propsheet.c \
 	rebar.c \
 	smoothscroll.c \
+	static.c \
 	status.c \
 	string.c \
+	syslink.c \
 	tab.c \
+	taskdialog.c \
 	toolbar.c \
 	tooltips.c \
 	trackbar.c \
 	treeview.c \
-	updown.c \
-	version.rc
+	updown.c
