diff --git a/dlls/krnl386.exe16/wowthunk.c b/dlls/krnl386.exe16/wowthunk.c
--- a/dlls/krnl386.exe16/wowthunk.c
+++ b/dlls/krnl386.exe16/wowthunk.c
@@ -73,12 +73,12 @@
     /* allocate the code selector for CallTo16 routines */
     WORD codesel = SELECTOR_AllocBlock( __wine_call16_start,
                                         (BYTE *)(&CallTo16_TebSelector + 1) - __wine_call16_start,
-                                        code32_segment );
+                                        LDT_FLAGS_CODE | LDT_FLAGS_32BIT );
 
     cbclient_selector = SELECTOR_AllocBlock( cbclient_ret, cbclient_ret_end - cbclient_ret,
-                                             code32_segment );
+                                             LDT_FLAGS_CODE | LDT_FLAGS_32BIT );
     cbclientex_selector = SELECTOR_AllocBlock( cbclientex_ret, cbclientex_ret_end - cbclientex_ret,
-                                               code32_segment );
+                                               LDT_FLAGS_CODE | LDT_FLAGS_32BIT );
     if (!codesel || !cbclient_selector || !cbclientex_selector)
         return FALSE;
 
@@ -170,7 +170,7 @@
                 *--stack = context->SegCs;
                 *--stack = context->Eip;
 
-                if (!ldt_is_32bit(context->SegSs))
+                if (!IS_SELECTOR_32BIT(context->SegSs))
                     context->Esp = MAKELONG( LOWORD(context->Esp - 2*sizeof(WORD)),
                                              HIWORD(context->Esp) );
                 else
@@ -460,9 +460,9 @@
         *((SEGPTR *)stack) = call16_ret_addr;
         cbArgs += sizeof(SEGPTR);
 
-        if (!(dwFlags & WCB16_INTERRUPT)) _EnterWin16Lock();
+        _EnterWin16Lock();
         wine_call_to_16_regs( context, cbArgs, call16_handler );
-        if (!(dwFlags & WCB16_INTERRUPT)) _LeaveWin16Lock();
+        _LeaveWin16Lock();
 
         if (TRACE_ON(relay))
         {
@@ -500,10 +500,10 @@
          * the callee to do so; after the routine has returned, the 16-bit
          * stack pointer is always reset to the position it had before.
          */
-        if (!(dwFlags & WCB16_INTERRUPT)) _EnterWin16Lock();
+        _EnterWin16Lock();
         ret = wine_call_to_16( (FARPROC16)vpfn16, cbArgs, call16_handler );
         if (pdwRetCode) *pdwRetCode = ret;
-        if (!(dwFlags & WCB16_INTERRUPT)) _LeaveWin16Lock();
+        _LeaveWin16Lock();
 
         if (TRACE_ON(relay))
         {
