diff --git a/dlls/rometadata/main.c b/dlls/rometadata/main.c
--- a/dlls/rometadata/main.c
+++ b/dlls/rometadata/main.c
@@ -21,10 +21,7 @@
 #include "objbase.h"
 #include "cor.h"
 #include "rometadata.h"
-#include "rometadataapi.h"
 #include "wine/debug.h"
-
-#include "rometadatapriv.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(rometadata);
 
@@ -90,35 +87,16 @@
 static HRESULT WINAPI MetaDataDispenser_OpenScope(IMetaDataDispenserEx *iface, const WCHAR *scope,
                                                   DWORD open_flags, REFIID riid, IUnknown **obj)
 {
-    IMetaDataTables *tables;
-    HRESULT hr;
-
-    FIXME("%p %s %lx %s %p semi-stub!\n", iface, debugstr_w(scope), open_flags, debugstr_guid(riid), obj);
-
-    *obj = NULL;
-    hr = IMetaDataTables_create_from_file(scope, &tables);
-    if (FAILED(hr))
-        return hr;
-
-    hr = IMetaDataTables_QueryInterface(tables, riid, (void **)obj);
-    IMetaDataTables_Release(tables);
-    return hr;
+    FIXME("%p %s %lx %s %p\n", iface, debugstr_w(scope), open_flags, debugstr_guid(riid), obj);
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI MetaDataDispenser_OpenScopeOnMemory(IMetaDataDispenserEx *iface, const void *data,
                                                           ULONG data_size, DWORD open_flags, REFIID riid,
                                                           IUnknown **obj)
 {
-    IMetaDataTables *tables;
-    HRESULT hr;
-
-    FIXME("%p %p %lu %lx %s %p semi-stub!\n", iface, data, data_size, open_flags, debugstr_guid(riid), obj);
-
-    *obj = NULL;
-    if (FAILED((hr = IMetaDataTables_create_from_data(data, data_size, &tables)))) return hr;
-    hr = IMetaDataTables_QueryInterface(tables, riid, (void **)obj);
-    IMetaDataTables_Release(tables);
-    return hr;
+    FIXME("%p %p %lu %lx %s %p\n", iface, data, data_size, open_flags, debugstr_guid(riid), obj);
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI MetaDataDispenser_SetOption(IMetaDataDispenserEx *iface, REFGUID option_id, const VARIANT *value)
