diff --git a/dlls/uiautomationcore/tests/uiautomation.c b/dlls/uiautomationcore/tests/uiautomation.c
--- a/dlls/uiautomationcore/tests/uiautomation.c
+++ b/dlls/uiautomationcore/tests/uiautomation.c
@@ -10035,14 +10035,14 @@
 
         SafeArrayDestroy(out_req);
         SysFreeString(tree_struct);
-        VariantClear(&prop_cond.Value);
+        VariantClear(&v);
 
         /* Same values, except we're short by one element. */
         V_VT(&v) = VT_I4 | VT_ARRAY;
         V_ARRAY(&v) = SafeArrayCreateVector(VT_I4, 0, ARRAY_SIZE(uia_i4_arr_prop_val) - 1);
 
         for (i = 0; i < ARRAY_SIZE(uia_i4_arr_prop_val) - 1; i++)
-            SafeArrayPutElement(V_ARRAY(&v), &i, (void *)&uia_i4_arr_prop_val[i]);
+            SafeArrayPutElement(V_ARRAY(&prop_cond.Value), &i, (void *)&uia_i4_arr_prop_val[i]);
 
         set_property_condition(&prop_cond, UIA_OutlineColorPropertyId, &v, PropertyConditionFlags_None);
         set_cache_request(&cache_req, (struct UiaCondition *)&prop_cond, TreeScope_Element, NULL, 0, NULL, 0,
@@ -10057,7 +10057,7 @@
         ok(!wcscmp(tree_struct, L""), "tree structure %s\n", debugstr_w(tree_struct));
 
         SysFreeString(tree_struct);
-        VariantClear(&prop_cond.Value);
+        VariantClear(&v);
     }
     else
         win_skip("UIA_OutlineColorPropertyId unavailable, skipping property condition tests for it.\n");
@@ -10476,7 +10476,6 @@
     { &Provider, PROV_GET_PROVIDER_OPTIONS, METHOD_OPTIONAL },
     { &Provider_nc, PROV_GET_PROVIDER_OPTIONS, METHOD_OPTIONAL },
     { &Provider_hwnd, PROV_GET_PROVIDER_OPTIONS, METHOD_OPTIONAL },
-    { 0 },
 };
 
 static const struct prov_method_sequence nav_seq10[] = {
@@ -10555,7 +10554,6 @@
 static const struct prov_method_sequence nav_seq15[] = {
     NODE_CREATE_SEQ(&Provider_child2_child_child),
     { &Provider_child2_child_child, PROV_GET_PROPERTY_VALUE }, /* UIA_ProviderDescriptionPropertyId */
-    { 0 },
 };
 
 static const struct prov_method_sequence nav_seq16[] = {
@@ -18629,7 +18627,7 @@
     memset(&startup, 0, sizeof(startup));
     startup.cb = sizeof(startup);
     CreateProcessA(NULL, cmdline, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &proc);
-    wait_child_process(&proc);
+    wait_child_process(proc.hProcess);
 }
 
 START_TEST(uiautomation)
