diff --git a/dlls/dsound/mixer.c b/dlls/dsound/mixer.c
--- a/dlls/dsound/mixer.c
+++ b/dlls/dsound/mixer.c
@@ -117,10 +117,15 @@
 	 * sample in the secondary buffer. firgain specifies what
 	 * to multiply the FIR output by in order to attenuate it correctly.
 	 */
-        if (dsb->freqAdjustNum > dsb->freqAdjustDen)
+	if (dsb->freqAdjustNum / dsb->freqAdjustDen > 0) {
+		/**
+		 * Yes, round it a bit to make sure that the
+		 * linear interpolation factor never changes.
+		 */
 		dsb->firstep = fir_step * dsb->freqAdjustDen / dsb->freqAdjustNum;
-        else
+	} else {
 		dsb->firstep = fir_step;
+	}
 	dsb->firgain = (float)dsb->firstep / fir_step;
 
 	/* calculate the 10ms write lead */
@@ -543,7 +548,8 @@
 	}
 
 	/* check for notification positions */
-	if (dsb->dsbd.dwFlags & DSBCAPS_CTRLPOSITIONNOTIFY) {
+	if (dsb->dsbd.dwFlags & DSBCAPS_CTRLPOSITIONNOTIFY &&
+	    dsb->state != STATE_STARTING) {
 		INT ilen = DSOUND_BufPtrDiff(dsb->buflen, dsb->sec_mixpos, oldpos);
 		DSOUND_CheckEvent(dsb, oldpos, ilen);
 	}
@@ -567,7 +573,22 @@
 	DWORD primary_done = 0;
 
 	TRACE("(%p, frames=%ld)\n",dsb,frames);
-	TRACE("looping=%ld\n", dsb->playflags);
+	TRACE("looping=%ld, leadin=%ld\n", dsb->playflags, dsb->leadin);
+
+	/* If leading in, only mix about 20 ms, and 'skip' mixing the rest, for more fluid pointer advancement */
+	/* FIXME: Is this needed? */
+	if (dsb->leadin && dsb->state == STATE_STARTING) {
+		if (frames > 2 * dsb->device->frag_frames) {
+			primary_done = frames - 2 * dsb->device->frag_frames;
+			frames = 2 * dsb->device->frag_frames;
+			dsb->sec_mixpos += primary_done *
+				dsb->pwfx->nBlockAlign * dsb->freqAdjustNum / dsb->freqAdjustDen;
+		}
+	}
+
+	dsb->leadin = FALSE;
+
+	TRACE("frames (primary) = %li\n", frames);
 
 	/* First try to mix to the end of the buffer if possible
 	 * Theoretically it would allow for better optimization
