diff --git a/dlls/winex11.drv/desktop.c b/dlls/winex11.drv/desktop.c
--- a/dlls/winex11.drv/desktop.c
+++ b/dlls/winex11.drv/desktop.c
@@ -35,7 +35,8 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(x11drv);
 
-static RECT host_primary_rect;
+static RECT *host_monitor_rects;
+static int host_monitor_rect_count;
 
 #define _NET_WM_STATE_REMOVE 0
 #define _NET_WM_STATE_ADD 1
@@ -51,9 +52,11 @@
  *
  * Setup the desktop when not using the root window.
  */
-void X11DRV_init_desktop( Window win, unsigned int width, unsigned int height )
+void X11DRV_init_desktop( Window win )
 {
-    host_primary_rect = get_host_primary_monitor_rect();
+    if (host_monitor_rects) free( host_monitor_rects );
+    if (!get_host_monitor_rects( &host_monitor_rects, &host_monitor_rect_count ))
+        ERR("Failed to get host monitor rectangle.\n");
     root_window = win;
     managed_mode = FALSE;  /* no managed windows in desktop mode */
 }
@@ -91,13 +94,23 @@
     x11drv_xinput2_enable( display, win );
     XFlush( display );
 
-    X11DRV_init_desktop( win, width, height );
+    X11DRV_init_desktop( win );
     return TRUE;
 }
 
 BOOL is_desktop_fullscreen(void)
 {
-    RECT primary_rect = NtUserGetPrimaryMonitorRect();
-    return (primary_rect.right - primary_rect.left == host_primary_rect.right - host_primary_rect.left &&
-            primary_rect.bottom - primary_rect.top == host_primary_rect.bottom - host_primary_rect.top);
+    Display *display = thread_display();
+    unsigned int width, height, border, depth;
+    int x, y, i;
+    Window root;
+    RECT rect;
+
+    XGetGeometry( display, root_window, &root, &x, &y, &width, &height, &border, &depth );
+    SetRect( &rect, x, y, x + width, y + height );
+
+    for (i = 0; i < host_monitor_rect_count; i++)
+        if (EqualRect( &host_monitor_rects[i], &rect) ) return TRUE;
+
+    return FALSE;
 }
