diff --git a/dlls/shell32/tests/ebrowser.c b/dlls/shell32/tests/ebrowser.c
--- a/dlls/shell32/tests/ebrowser.c
+++ b/dlls/shell32/tests/ebrowser.c
@@ -988,6 +988,7 @@
     HDWP hdwp;
     RECT rc;
     HRESULT hr;
+    static const WCHAR winetest[] = {'W','i','n','e','T','e','s','t',0};
 
     ebrowser_instantiate(&peb);
     ebrowser_initialize(peb);
@@ -1089,7 +1090,7 @@
     /* SetPropertyBag */
     hr = IExplorerBrowser_SetPropertyBag(peb, NULL);
     ok(hr == E_INVALIDARG, "Got 0x%08lx\n", hr);
-    hr = IExplorerBrowser_SetPropertyBag(peb, L"WineTest");
+    hr = IExplorerBrowser_SetPropertyBag(peb, winetest);
     ok(hr == S_OK, "Got 0x%08lx\n", hr);
 
     /* TODO: Test after browsing somewhere. */
@@ -1280,7 +1281,8 @@
     LONG lres;
     WCHAR current_path[MAX_PATH];
     WCHAR child_path[MAX_PATH];
-    static WCHAR testfolderW[] = L"winetestfolder";
+    static const WCHAR testfolderW[] =
+        {'w','i','n','e','t','e','s','t','f','o','l','d','e','r','\0'};
 
     ok(pSHCreateShellItem != NULL, "pSHCreateShellItem unexpectedly missing.\n");
 
@@ -1360,7 +1362,8 @@
 
         hr = IFolderView_GetFolder(pfv, &IID_IShellFolder, (void**)&psf);
         ok(hr == S_OK, "Got 0x%08lx\n", hr);
-        hr = IShellFolder_ParseDisplayName(psf, NULL, NULL, testfolderW, NULL, &pidl_relative, NULL);
+        hr = IShellFolder_ParseDisplayName(psf, NULL, NULL, (LPWSTR)testfolderW,
+                                           NULL, &pidl_relative, NULL);
         ok(hr == S_OK, "Got 0x%08lx\n", hr);
 
         /* Browsing to another location here before using the
@@ -1673,12 +1676,13 @@
 static void setup_window(void)
 {
     WNDCLASSW wc;
+    static const WCHAR ebtestW[] = {'e','b','t','e','s','t',0};
 
     ZeroMemory(&wc, sizeof(WNDCLASSW));
     wc.lpfnWndProc      = DefWindowProcW;
-    wc.lpszClassName    = L"ebtest";
+    wc.lpszClassName    = ebtestW;
     RegisterClassW(&wc);
-    hwnd = CreateWindowExW(0, L"ebtest", NULL, 0,
+    hwnd = CreateWindowExW(0, ebtestW, NULL, 0,
                            0, 0, 500, 500,
                            NULL, 0, 0, NULL);
     ok(hwnd != NULL, "Failed to create window for tests.\n");
