diff --git a/dlls/msvcp90/msvcp90.h b/dlls/msvcp90/msvcp90.h
--- a/dlls/msvcp90/msvcp90.h
+++ b/dlls/msvcp90/msvcp90.h
@@ -27,8 +27,8 @@
 #define ALIGNED_SIZE(size, alignment) (((size)+((alignment)-1))/(alignment)*(alignment))
 
 #if _MSVCP_VER >= 100
-typedef INT64 streamoff;
-typedef INT64 streamsize;
+typedef __int64 DECLSPEC_ALIGN(8) streamoff;
+typedef __int64 DECLSPEC_ALIGN(8) streamsize;
 #else
 typedef SSIZE_T streamoff;
 typedef SSIZE_T streamsize;
@@ -36,9 +36,10 @@
 
 void __cdecl _invalid_parameter_noinfo(void);
 BOOL __cdecl __uncaught_exception(void);
-
+int __cdecl _callnewh(size_t);
+
+void* __cdecl operator_new(size_t);
 void __cdecl operator_delete(void*);
-void* __cdecl operator_new(size_t) __WINE_ALLOC_SIZE(1) __WINE_DEALLOC(operator_delete) __WINE_MALLOC;
 extern void* (__cdecl *MSVCRT_set_new_handler)(void*);
 
 #if _MSVCP_VER >= 110
@@ -159,8 +160,8 @@
 basic_string_wchar* __thiscall MSVCP_basic_string_wchar_append_ch(basic_string_wchar*, wchar_t);
 size_t __thiscall MSVCP_basic_string_wchar_length(const basic_string_wchar*);
 
+char* __thiscall MSVCP_allocator_char_allocate(void*, size_t);
 void __thiscall MSVCP_allocator_char_deallocate(void*, char*, size_t);
-char* __thiscall MSVCP_allocator_char_allocate(void*, size_t) __WINE_ALLOC_SIZE(2) __WINE_DEALLOC(MSVCP_allocator_char_deallocate, 2) __WINE_MALLOC;
 size_t __thiscall MSVCP_allocator_char_max_size(const void*);
 wchar_t* __thiscall MSVCP_allocator_wchar_allocate(void*, size_t);
 void __thiscall MSVCP_allocator_wchar_deallocate(void*, wchar_t*, size_t);
@@ -193,11 +194,7 @@
 /* class locale::facet */
 typedef struct {
     const vtable_ptr *vtable;
-#if _MSVCP_VER >= 110
-    unsigned int refs;
-#else
     size_t refs;
-#endif
 } locale_facet;
 
 typedef enum {
@@ -733,7 +730,6 @@
 #define hypotf( x, y ) ((float)hypot( (double)(x), (double)(y) ))
 #endif
 
-void CDECL __DestructExceptionObject(EXCEPTION_RECORD*);
 void WINAPI DECLSPEC_NORETURN _CxxThrowException(void*,const cxx_exception_type*);
 void __cdecl DECLSPEC_NORETURN _Xinvalid_argument(const char*);
 void __cdecl DECLSPEC_NORETURN _Xlength_error(const char*);
