diff --git a/dlls/bcrypt/bcrypt_internal.h b/dlls/bcrypt/bcrypt_internal.h
--- a/dlls/bcrypt/bcrypt_internal.h
+++ b/dlls/bcrypt/bcrypt_internal.h
@@ -62,14 +62,12 @@
 
     /* secret agreement */
     ALG_ID_DH,
-    ALG_ID_ECDH,
     ALG_ID_ECDH_P256,
     ALG_ID_ECDH_P384,
     ALG_ID_ECDH_P521,
 
     /* signature */
     ALG_ID_RSA_SIGN,
-    ALG_ID_ECDSA,
     ALG_ID_ECDSA_P256,
     ALG_ID_ECDSA_P384,
     ALG_ID_ECDSA_P521,
@@ -91,21 +89,12 @@
     CHAIN_MODE_GCM,
 };
 
-enum ecc_curve_id
-{
-    ECC_CURVE_NONE,
-    ECC_CURVE_P256R1,
-    ECC_CURVE_P384R1,
-    ECC_CURVE_P521R1,
-};
-
 struct algorithm
 {
-    struct object     hdr;
-    enum alg_id       id;
-    enum chain_mode   mode;
-    unsigned          flags;
-    enum ecc_curve_id curve_id;
+    struct object   hdr;
+    enum alg_id     id;
+    enum chain_mode mode;
+    unsigned        flags;
 };
 
 struct key_symmetric
@@ -121,11 +110,11 @@
 
 #define KEY_FLAG_LEGACY_DSA_V2  0x00000001
 #define KEY_FLAG_FINALIZED      0x00000002
+#define KEY_FLAG_DH_PARAMS_SET  0x00000004
 
 struct key_asymmetric
 {
-    ULONG             bitlen;     /* key strength for ECC keys */
-    enum ecc_curve_id curve_id;
+    ULONG             bitlen;     /* ignored for ECC keys */
     unsigned          flags;
     DSSSEED           dss_seed;
 };
