diff --git a/dlls/secur32/negotiate.c b/dlls/secur32/negotiate.c
--- a/dlls/secur32/negotiate.c
+++ b/dlls/secur32/negotiate.c
@@ -47,7 +47,7 @@
 };
 
 #define WINE_NO_CACHED_CREDENTIALS 0x10000000
-#define NEGO_MAX_TOKEN 48256
+#define NEGO_MAX_TOKEN 12000
 
 static WCHAR nego_name_W[] = {'N','e','g','o','t','i','a','t','e',0};
 static char nego_name_A[] = "Negotiate";
@@ -238,14 +238,8 @@
         if ((ret == SEC_E_OK || ret == SEC_I_CONTINUE_NEEDED) && new_context)
         {
             ctxt->ntlm = NULL;
-            ctxt->user_ntlm = NULL;
             *new_context = (LSA_SEC_HANDLE)ctxt;
             if (new_ctxt == ctxt) new_ctxt = NULL;
-        }
-        else
-        {
-            ctxt->krb = NULL;
-            ctxt->user_krb = NULL;
         }
     }
 
@@ -257,7 +251,6 @@
         if ((ret == SEC_E_OK || ret == SEC_I_CONTINUE_NEEDED) && new_context)
         {
             ctxt->krb = NULL;
-            ctxt->user_krb = NULL;
             *new_context = (LSA_SEC_HANDLE)ctxt;
             if (new_ctxt == ctxt) new_ctxt = NULL;
         }
@@ -301,14 +294,8 @@
         if ((ret == SEC_E_OK || ret == SEC_I_CONTINUE_NEEDED) && new_context)
         {
             ctxt->ntlm = NULL;
-            ctxt->user_ntlm = NULL;
             *new_context = (LSA_SEC_HANDLE)ctxt;
             if (new_ctxt == ctxt) new_ctxt = NULL;
-        }
-        else
-        {
-            ctxt->krb = NULL;
-            ctxt->user_krb = NULL;
         }
     }
 
@@ -321,7 +308,6 @@
         if ((ret == SEC_E_OK || ret == SEC_I_CONTINUE_NEEDED) && new_context)
         {
             ctxt->krb = NULL;
-            ctxt->user_krb = NULL;
             *new_context = (LSA_SEC_HANDLE)ctxt;
             if (new_ctxt == ctxt) new_ctxt = NULL;
         }
