diff --git a/dlls/crypt32/regstore.c b/dlls/crypt32/regstore.c
--- a/dlls/crypt32/regstore.c
+++ b/dlls/crypt32/regstore.c
@@ -43,7 +43,6 @@
     struct list      certsToDelete;
     struct list      crlsToDelete;
     struct list      ctlsToDelete;
-    HANDLE           key_modified_event;
 } WINE_REGSTOREINFO;
 
 void CRYPT_HashToStr(const BYTE *hash, LPWSTR asciiHash)
@@ -325,7 +324,6 @@
 
     CRYPT_RegFlushStore(store, FALSE);
     RegCloseKey(store->key);
-    CloseHandle(store->key_modified_event);
     store->cs.DebugInfo->Spare[0] = 0;
     DeleteCriticalSection(&store->cs);
     CryptMemFree(store);
@@ -463,22 +461,13 @@
     {
     case CERT_STORE_CTRL_RESYNC:
     {
-        HCERTSTORE memStore;
-        DWORD ret;
-
-        EnterCriticalSection(&store->cs);
+        HCERTSTORE memStore = CertOpenStore(CERT_STORE_PROV_MEMORY, 0, 0,
+         CERT_STORE_CREATE_NEW_FLAG, NULL);
+
         CRYPT_RegFlushStore(store, FALSE);
-        if ((ret = WaitForSingleObject(store->key_modified_event, 0)) != WAIT_TIMEOUT)
-        {
-            memStore = CertOpenStore(CERT_STORE_PROV_MEMORY, 0, 0, CERT_STORE_CREATE_NEW_FLAG, NULL);
-            if (RegNotifyChangeKeyValue(store->key, TRUE, REG_NOTIFY_CHANGE_NAME | REG_NOTIFY_CHANGE_LAST_SET,
-                store->key_modified_event, TRUE))
-                ERR("RegNotifyChangeKeyValue failed.\n");
-            CRYPT_RegReadFromReg(store->key, memStore, CERT_STORE_ADD_REPLACE_EXISTING);
-            I_CertUpdateStore(store->memStore, memStore, 0, 0);
-            CertCloseStore(memStore, 0);
-        }
-        LeaveCriticalSection(&store->cs);
+        CRYPT_RegReadFromReg(store->key, memStore, CERT_STORE_ADD_REPLACE_EXISTING);
+        I_CertUpdateStore(store->memStore, memStore, 0, 0);
+        CertCloseStore(memStore, 0);
         break;
     }
     case CERT_STORE_CTRL_COMMIT:
@@ -564,10 +553,6 @@
                     list_init(&regInfo->certsToDelete);
                     list_init(&regInfo->crlsToDelete);
                     list_init(&regInfo->ctlsToDelete);
-                    regInfo->key_modified_event = CreateEventW(NULL, FALSE, FALSE, NULL);
-                    if (RegNotifyChangeKeyValue(regInfo->key, TRUE, REG_NOTIFY_CHANGE_NAME | REG_NOTIFY_CHANGE_LAST_SET,
-                        regInfo->key_modified_event, TRUE))
-                        ERR("RegNotifyChangeKeyValue failed.\n");
                     CRYPT_RegReadFromReg(regInfo->key, regInfo->memStore, CERT_STORE_ADD_ALWAYS);
                     regInfo->dirty = FALSE;
                     provInfo.cbSize = sizeof(provInfo);
