diff --git a/dlls/wow64/file.c b/dlls/wow64/file.c
--- a/dlls/wow64/file.c
+++ b/dlls/wow64/file.c
@@ -37,8 +37,7 @@
 
 static inline NTSTATUS get_file_id( HANDLE handle, FILE_OBJECTID_BUFFER *id )
 {
-    IO_STATUS_BLOCK32 io32;
-    IO_STATUS_BLOCK io = { .Pointer = &io32 };
+    IO_STATUS_BLOCK io;
 
     return NtFsControlFile( handle, 0, NULL, NULL, &io, FSCTL_GET_OBJECT_ID, NULL, 0, id, sizeof(*id) );
 }
@@ -752,11 +751,7 @@
 
     NTSTATUS status;
     ULONG i;
-    FILE_IO_COMPLETION_INFORMATION *info;
-
-    if (!count) return STATUS_INVALID_PARAMETER;
-
-    info = Wow64AllocateTemp( count * sizeof(*info) );
+    FILE_IO_COMPLETION_INFORMATION *info = Wow64AllocateTemp( count * sizeof(*info) );
 
     status = NtRemoveIoCompletionEx( handle, info, count, written, timeout, alertable );
     for (i = 0; i < *written; i++)
@@ -956,3 +951,32 @@
     put_iosb( io32, &io );
     return status;
 }
+
+
+/**********************************************************************
+ *           wow64_wine_nt_to_unix_file_name
+ */
+NTSTATUS WINAPI wow64_wine_nt_to_unix_file_name( UINT *args )
+{
+    OBJECT_ATTRIBUTES32 *attr32 = get_ptr( &args );
+    char *nameA = get_ptr( &args );
+    ULONG *size = get_ptr( &args );
+    UINT disposition = get_ulong( &args );
+
+    struct object_attr64 attr;
+
+    return wine_nt_to_unix_file_name( objattr_32to64_redirect( &attr, attr32 ), nameA, size, disposition );
+}
+
+
+/**********************************************************************
+ *           wow64_wine_unix_to_nt_file_name
+ */
+NTSTATUS WINAPI wow64_wine_unix_to_nt_file_name( UINT *args )
+{
+    const char *name = get_ptr( &args );
+    WCHAR *buffer = get_ptr( &args );
+    ULONG *size = get_ptr( &args );
+
+    return wine_unix_to_nt_file_name( name, buffer, size );
+}
