diff --git a/dlls/rsaenh/implglue.h b/dlls/rsaenh/implglue.h
--- a/dlls/rsaenh/implglue.h
+++ b/dlls/rsaenh/implglue.h
@@ -24,28 +24,24 @@
 #ifndef __WINE_IMPLGLUE_H
 #define __WINE_IMPLGLUE_H
 
+#include "bcrypt.h"
 #include "tomcrypt.h"
 
 #define RSAENH_MAX_HASH_SIZE        104
 
 typedef union tagKEY_CONTEXT {
-    symmetric_key key;
-    prng_state prng;
+    rc2_key rc2;
+    des_key des;
+    des3_key des3;
+    aes_key aes;
+    prng_state rc4;
     rsa_key rsa;
 } KEY_CONTEXT;
 
-extern prng_state prng;
-extern int wprng;
-
-struct hash
-{
-    const struct ltc_hash_descriptor *desc;
-    hash_state state;
-};
-
-BOOL init_hash_impl(ALG_ID algid, struct hash *hash);
-BOOL update_hash_impl(struct hash *hash, const BYTE *data, DWORD len);
-BOOL finalize_hash_impl(struct hash *hash, BYTE *hash_value, DWORD hash_size);
+BOOL init_hash_impl(ALG_ID aiAlgid, BCRYPT_HASH_HANDLE *hash_handle);
+BOOL update_hash_impl(BCRYPT_HASH_HANDLE hash_handle, const BYTE *pbData, DWORD dwDataLen);
+BOOL finalize_hash_impl(BCRYPT_HASH_HANDLE hash_handle, BYTE *hash_value, DWORD hash_size);
+BOOL duplicate_hash_impl(BCRYPT_HASH_HANDLE src_hash_handle, BCRYPT_HASH_HANDLE *dest_hash_handle);
 
 BOOL new_key_impl(ALG_ID aiAlgid, KEY_CONTEXT *pKeyContext, DWORD dwKeyLen);
 BOOL free_key_impl(ALG_ID aiAlgid, KEY_CONTEXT *pKeyContext);
@@ -56,9 +52,7 @@
 
 /* dwKeySpec is optional for symmetric key algorithms */
 BOOL encrypt_block_impl(ALG_ID aiAlgid, DWORD dwKeySpec, KEY_CONTEXT *pKeyContext, const BYTE *pbIn,
-                        BYTE *pbOut);
-BOOL decrypt_block_impl(ALG_ID aiAlgid, DWORD dwKeySpec, KEY_CONTEXT *pKeyContext, const BYTE *pbIn,
-                        BYTE *pbOut);
+                        BYTE *pbOut, DWORD enc);
 BOOL encrypt_stream_impl(ALG_ID aiAlgid, KEY_CONTEXT *pKeyContext, BYTE *pbInOut, DWORD dwLen);
 
 BOOL export_public_key_impl(BYTE *pbDest, const KEY_CONTEXT *pKeyContext, DWORD dwKeyLen,
@@ -70,4 +64,6 @@
 BOOL import_private_key_impl(const BYTE* pbSrc, KEY_CONTEXT *pKeyContext, DWORD dwKeyLen,
                              DWORD dwDataLen, DWORD dwPubExp);
 
+BOOL gen_rand_impl(BYTE *pbBuffer, DWORD dwLen);
+
 #endif /* __WINE_IMPLGLUE_H */
