diff --git a/dlls/cryptsp/cryptsp.spec b/dlls/cryptsp/cryptsp.spec
--- a/dlls/cryptsp/cryptsp.spec
+++ b/dlls/cryptsp/cryptsp.spec
@@ -1,4 +1,4 @@
-@ stdcall CheckSignatureInFile(str)
+@ stub CheckSignatureInFile
 @ stdcall -import CryptAcquireContextA(ptr str str long long)
 @ stdcall -import CryptAcquireContextW(ptr wstr wstr long long)
 @ stdcall -import CryptContextAddRef(long ptr long)
@@ -38,28 +38,28 @@
 @ stdcall -import CryptSignHashW(long long wstr long ptr ptr)
 @ stdcall -import CryptVerifySignatureA(long ptr long long str long)
 @ stdcall -import CryptVerifySignatureW(long ptr long long wstr long)
-@ stdcall SystemFunction006(ptr ptr)
-@ stdcall SystemFunction007(ptr ptr)
-@ stdcall SystemFunction008(ptr ptr ptr)
-@ stdcall SystemFunction009(ptr ptr ptr) SystemFunction008
-@ stdcall SystemFunction010(ptr ptr ptr)
-@ stdcall SystemFunction011(ptr ptr ptr) SystemFunction010
-@ stdcall SystemFunction012(ptr ptr ptr)
-@ stdcall SystemFunction013(ptr ptr ptr)
-@ stdcall SystemFunction014(ptr ptr ptr) SystemFunction012
-@ stdcall SystemFunction015(ptr ptr ptr) SystemFunction013
-@ stdcall SystemFunction016(ptr ptr ptr) SystemFunction012
-@ stdcall SystemFunction018(ptr ptr ptr) SystemFunction012
-@ stdcall SystemFunction020(ptr ptr ptr) SystemFunction012
-@ stdcall SystemFunction021(ptr ptr ptr) SystemFunction013
-@ stdcall SystemFunction022(ptr ptr ptr) SystemFunction012
-@ stdcall SystemFunction023(ptr ptr ptr) SystemFunction013
-@ stdcall SystemFunction024(ptr ptr ptr)
-@ stdcall SystemFunction025(ptr ptr ptr)
-@ stdcall SystemFunction026(ptr ptr ptr) SystemFunction024
-@ stdcall SystemFunction027(ptr ptr ptr) SystemFunction025
-@ stdcall SystemFunction030(ptr ptr)
-@ stdcall SystemFunction031(ptr ptr) SystemFunction030
-@ stdcall SystemFunction032(ptr ptr)
+@ stdcall -import SystemFunction006(ptr ptr)
+@ stdcall -import SystemFunction007(ptr ptr)
+@ stdcall -import SystemFunction008(ptr ptr ptr)
+@ stdcall -import SystemFunction009(ptr ptr ptr)
+@ stdcall -import SystemFunction010(ptr ptr ptr)
+@ stdcall -import SystemFunction011(ptr ptr ptr)
+@ stdcall -import SystemFunction012(ptr ptr ptr)
+@ stdcall -import SystemFunction013(ptr ptr ptr)
+@ stdcall -import SystemFunction014(ptr ptr ptr)
+@ stdcall -import SystemFunction015(ptr ptr ptr)
+@ stdcall -import SystemFunction016(ptr ptr ptr)
+@ stdcall -import SystemFunction018(ptr ptr ptr)
+@ stdcall -import SystemFunction020(ptr ptr ptr)
+@ stdcall -import SystemFunction021(ptr ptr ptr)
+@ stdcall -import SystemFunction022(ptr ptr ptr)
+@ stdcall -import SystemFunction023(ptr ptr ptr)
+@ stdcall -import SystemFunction024(ptr ptr ptr)
+@ stdcall -import SystemFunction025(ptr ptr ptr)
+@ stdcall -import SystemFunction026(ptr ptr ptr)
+@ stdcall -import SystemFunction027(ptr ptr ptr)
+@ stdcall -import SystemFunction030(ptr ptr)
+@ stdcall -import SystemFunction031(ptr ptr)
+@ stdcall -import SystemFunction032(ptr ptr)
 @ stub SystemFunction033
-@ stdcall SystemFunction035(str) CheckSignatureInFile
+@ stdcall -import SystemFunction035(str)
