diff --git a/dlls/dwrite/dwrite_private.h b/dlls/dwrite/dwrite_private.h
--- a/dlls/dwrite/dwrite_private.h
+++ b/dlls/dwrite/dwrite_private.h
@@ -18,8 +18,6 @@
 
 #ifndef __WINE_DWRITE_PRIVATE_H
 #define __WINE_DWRITE_PRIVATE_H
-
-#include <stdbool.h>
 
 #include "dwrite_3.h"
 #include "d2d1.h"
@@ -340,9 +338,10 @@
 extern IDWriteFontFileLoader *get_local_fontfile_loader(void);
 extern HRESULT create_fontface(const struct fontface_desc *desc, struct list *cached_list,
         IDWriteFontFace5 **fontface);
-extern HRESULT create_font_collection(IDWriteFactory7 *factory, IDWriteFontFileEnumerator *enumerator,
-        IDWriteFontCollection3 **collection);
+extern HRESULT create_font_collection(IDWriteFactory7 *factory, IDWriteFontFileEnumerator *enumerator, BOOL is_system,
+       IDWriteFontCollection3 **collection);
 extern HRESULT create_glyphrunanalysis(const struct glyphrunanalysis_desc*,IDWriteGlyphRunAnalysis**);
+extern BOOL    is_system_collection(IDWriteFontCollection*);
 extern HRESULT get_local_refkey(const WCHAR*,const FILETIME*,void**,UINT32*);
 extern HRESULT get_filestream_from_file(IDWriteFontFile*,IDWriteFontFileStream**);
 extern BOOL    is_face_type_supported(DWRITE_FONT_FACE_TYPE);
@@ -366,7 +365,6 @@
 extern void factory_detach_gdiinterop(IDWriteFactory7 *factory, IDWriteGdiInterop1 *interop);
 extern struct fontfacecached *factory_cache_fontface(IDWriteFactory7 *factory, struct list *fontfaces,
         IDWriteFontFace5 *fontface);
-extern IDWriteFontFile *get_fontfile_from_font(IDWriteFont *font);
 extern void    get_logfont_from_font(IDWriteFont*,LOGFONTW*);
 extern void    get_logfont_from_fontface(IDWriteFontFace*,LOGFONTW*);
 extern HRESULT get_fontsig_from_font(IDWriteFont*,FONTSIGNATURE*);
@@ -378,19 +376,12 @@
 extern HRESULT create_inmemory_fileloader(IDWriteInMemoryFontFileLoader **loader);
 extern HRESULT create_font_resource(IDWriteFactory7 *factory, IDWriteFontFile *file, UINT32 face_index,
         IDWriteFontResource **resource);
-extern HRESULT create_fontset_builder(IDWriteFactory7 *factory, BOOL is_system, IDWriteFontSetBuilder2 **ret);
+extern HRESULT create_fontset_builder(IDWriteFactory7 *factory, IDWriteFontSetBuilder2 **ret);
 extern HRESULT compute_glyph_origins(DWRITE_GLYPH_RUN const *run, DWRITE_MEASURING_MODE measuring_mode,
         D2D1_POINT_2F baseline_origin, DWRITE_MATRIX const *transform, D2D1_POINT_2F *origins);
 extern HRESULT create_font_collection_from_set(IDWriteFactory7 *factory, IDWriteFontSet *set,
         DWRITE_FONT_FAMILY_MODEL family_model, REFGUID riid, void **ret);
 extern HRESULT create_system_fontset(IDWriteFactory7 *factory, REFIID riid, void **obj);
-
-struct dwrite_fontset_entry;
-extern void release_fontset_entry(struct dwrite_fontset_entry *);
-extern HRESULT fontset_builder_get_entries(IDWriteFontSetBuilder2 *iface, struct dwrite_fontset_entry ***ret,
-        unsigned int *count);
-extern HRESULT fontset_create_from_set(IDWriteFactory7 *factory, struct dwrite_fontset_entry **src_entries,
-        unsigned int count, BOOL is_system, IDWriteFontSet **ret);
 
 struct dwrite_fontface;
 
@@ -542,7 +533,6 @@
 extern BOOL opentype_has_kerning_pairs(struct dwrite_fontface *fontface);
 extern HRESULT opentype_get_font_var_axis(const struct file_stream_desc *stream_desc, struct dwrite_var_axis **axis,
         unsigned int *axis_count);
-extern bool opentype_is_whitespace(unsigned int codepoint);
 
 struct dwrite_colorglyph {
     USHORT layer; /* [0, num_layers) index indicating current layer */
