diff --git a/dlls/mf/sar.c b/dlls/mf/sar.c
--- a/dlls/mf/sar.c
+++ b/dlls/mf/sar.c
@@ -1339,6 +1339,7 @@
 {
     struct queued_object *object;
     DWORD sample_len, sample_frames;
+    MFTIME time, clocktime, systime;
     HRESULT hr;
 
     if (FAILED(hr = IMFSample_GetTotalLength(sample, &sample_len)))
@@ -1346,15 +1347,33 @@
 
     sample_frames = sample_len / renderer->frame_size;
 
-    if (!(object = calloc(1, sizeof(*object))))
-        return E_OUTOFMEMORY;
-
-    object->type = OBJECT_TYPE_SAMPLE;
-    object->u.sample.sample = sample;
-    IMFSample_AddRef(object->u.sample.sample);
-
-    list_add_tail(&renderer->queue, &object->entry);
-    renderer->queued_frames += sample_frames;
+    if (FAILED(hr = IMFSample_GetSampleTime(sample, &time)))
+    {
+        WARN("Failed to get sample time, hr %#lx.\n", hr);
+        return hr;
+    }
+
+    if (!renderer->clock)
+        clocktime = time;
+    else if (FAILED(hr = IMFPresentationClock_GetCorrelatedTime(renderer->clock, 0, &clocktime, &systime)))
+    {
+        WARN("Failed to get clock time, hr %#lx.\n", hr);
+        return hr;
+    }
+
+    if (time < clocktime)
+        FIXME("Dropping sample %p, time %I64u, clocktime %I64u, systime %I64u.\n", sample, time, clocktime, systime);
+    else
+    {
+        if (!(object = calloc(1, sizeof(*object))))
+            return E_OUTOFMEMORY;
+
+        object->type = OBJECT_TYPE_SAMPLE;
+        object->u.sample.sample = sample;
+        IMFSample_AddRef(object->u.sample.sample);
+        list_add_tail(&renderer->queue, &object->entry);
+        renderer->queued_frames += sample_frames;
+    }
 
     return S_OK;
 }
@@ -1505,6 +1524,7 @@
 {
     static const GUID *required_attrs[] =
     {
+        &MF_MT_SUBTYPE,
         &MF_MT_AUDIO_SAMPLES_PER_SECOND,
         &MF_MT_AUDIO_NUM_CHANNELS,
         &MF_MT_AUDIO_BITS_PER_SAMPLE,
