diff --git a/dlls/mountmgr.sys/unixlib.c b/dlls/mountmgr.sys/unixlib.c
--- a/dlls/mountmgr.sys/unixlib.c
+++ b/dlls/mountmgr.sys/unixlib.c
@@ -491,10 +491,12 @@
 static NTSTATUS set_shell_folder( void *args )
 {
     const struct set_shell_folder_params *params = args;
+    const char *folder = params->folder;
+    const char *backup = params->backup;
     const char *link = params->link;
     struct stat st;
     const char *home;
-    char *folder = NULL, *backup = NULL, *homelink = NULL;
+    char *homelink = NULL;
     NTSTATUS status = STATUS_SUCCESS;
 
     if (link && (!strcmp( link, "$HOME" ) || !strncmp( link, "$HOME/", 6 )) && (home = getenv( "HOME" )))
@@ -510,12 +512,6 @@
         status = STATUS_OBJECT_NAME_NOT_FOUND;
         goto done;
     }
-
-    status = ntdll_get_unix_file_name( params->folder, &folder, FILE_OPEN_IF );
-    if (status == STATUS_NO_SUCH_FILE) status = STATUS_SUCCESS;
-    if (status) goto done;
-
-    if (params->create_backup) asprintf( &backup, "%s.backup", folder );
 
     if (!lstat( folder, &st )) /* move old folder/link out of the way */
     {
@@ -545,8 +541,6 @@
     }
 
 done:
-    free( backup );
-    free( folder );
     free( homelink );
     return status;
 }
@@ -554,17 +548,11 @@
 static NTSTATUS get_shell_folder( void *args )
 {
     const struct get_shell_folder_params *params = args;
-    char *folder = NULL;
-    NTSTATUS status = ntdll_get_unix_file_name( params->folder, &folder, FILE_OPEN );
-
-    if (!status)
-    {
-        int ret = readlink( folder, params->buffer, params->size - 1 );
-        free( folder );
-        if (ret < 0) return STATUS_OBJECT_NAME_NOT_FOUND;
-        params->buffer[ret] = 0;
-    }
-    return status;
+    int ret = readlink( params->folder, params->buffer, params->size - 1 );
+
+    if (ret < 0) return STATUS_OBJECT_NAME_NOT_FOUND;
+    params->buffer[ret] = 0;
+    return STATUS_SUCCESS;
 }
 
 const unixlib_entry_t __wine_unix_call_funcs[] =
