diff --git a/dlls/d2d1/bitmap.c b/dlls/d2d1/bitmap.c
--- a/dlls/d2d1/bitmap.c
+++ b/dlls/d2d1/bitmap.c
@@ -158,8 +158,7 @@
     ID3D11Device *device;
     D3D11_BOX box;
 
-    TRACE("iface %p, dst_point %s, bitmap %p, src_rect %s.\n", iface,
-            debug_d2d_point_2u(dst_point), bitmap, debug_d2d_rect_u(src_rect));
+    TRACE("iface %p, dst_point %p, bitmap %p, src_rect %p.\n", iface, dst_point, bitmap, src_rect);
 
     if (src_rect)
     {
@@ -185,36 +184,9 @@
 static HRESULT STDMETHODCALLTYPE d2d_bitmap_CopyFromRenderTarget(ID2D1Bitmap1 *iface,
         const D2D1_POINT_2U *dst_point, ID2D1RenderTarget *render_target, const D2D1_RECT_U *src_rect)
 {
-    ID2D1DeviceContext *device_context;
-    ID2D1Bitmap *target_bitmap = NULL;
-    ID2D1Image *target = NULL;
-    HRESULT hr;
-
-    TRACE("iface %p, dst_point %s, render_target %p, src_rect %s.\n", iface,
-            debug_d2d_point_2u(dst_point), render_target, debug_d2d_rect_u(src_rect));
-
-    if (FAILED(hr = ID2D1RenderTarget_QueryInterface(render_target, &IID_ID2D1DeviceContext,
-            (void **)&device_context)))
-    {
-        return hr;
-    }
-
-    ID2D1DeviceContext_GetTarget(device_context, &target);
-    ID2D1DeviceContext_Release(device_context);
-
-    if (target)
-    {
-        ID2D1Image_QueryInterface(target, &IID_ID2D1Bitmap, (void **)&target_bitmap);
-        ID2D1Image_Release(target);
-    }
-
-    if (!target_bitmap)
-        return D2DERR_RECREATE_TARGET;
-
-    hr = ID2D1Bitmap1_CopyFromBitmap(iface, dst_point, target_bitmap, src_rect);
-    ID2D1Bitmap_Release(target_bitmap);
-
-    return hr;
+    FIXME("iface %p, dst_point %p, render_target %p, src_rect %p stub!\n", iface, dst_point, render_target, src_rect);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_bitmap_CopyFromMemory(ID2D1Bitmap1 *iface,
@@ -225,8 +197,7 @@
     ID3D11Device *device;
     D3D11_BOX box;
 
-    TRACE("iface %p, dst_rect %s, src_data %p, pitch %u.\n", iface, debug_d2d_rect_u(dst_rect),
-            src_data, pitch);
+    TRACE("iface %p, dst_rect %p, src_data %p, pitch %u.\n", iface, dst_rect, src_data, pitch);
 
     if (dst_rect)
     {
