diff --git a/dlls/setupapi/tests/devinst.c b/dlls/setupapi/tests/devinst.c
--- a/dlls/setupapi/tests/devinst.c
+++ b/dlls/setupapi/tests/devinst.c
@@ -2142,9 +2142,11 @@
 
     SetLastError(0xdeadbeef);
     key = SetupDiOpenDevRegKey(set, &device, DICS_FLAG_GLOBAL, 0, DIREG_DRV, 0);
+todo_wine {
     ok(key == INVALID_HANDLE_VALUE, "Expected failure.\n");
     ok(GetLastError() == ERROR_INVALID_DATA || GetLastError() == ERROR_ACCESS_DENIED, /* win2k3 */
             "Got unexpected error %#lx.\n", GetLastError());
+}
 
     key = SetupDiOpenDevRegKey(set, &device, DICS_FLAG_GLOBAL, 0, DIREG_DRV, KEY_READ);
     ok(key != INVALID_HANDLE_VALUE, "Failed to open device key, error %#lx.\n", GetLastError());
@@ -5269,34 +5271,34 @@
         len = 0;
         proptype = DEVPROP_TYPE_EMPTY;
         ret = SetupDiGetDevicePropertyW(set, &devinfodata, &DEVPKEY_Device_DriverDate,  &proptype, (BYTE *)&filetime, sizeof(LARGE_INTEGER), &len, 0);
-        ok(ret, "getting DEVPKEY_Device_DriverDate failed with error %lu.\n", GetLastError());
-        ok(proptype == DEVPROP_TYPE_FILETIME, "got unexpected proptype %#lx.\n", proptype);
-        ok(len == sizeof(LARGE_INTEGER), "got unexpected length.\n");
-        ok((filetime.QuadPart % 864000000000) == 0, "returned value should not contain a time.\n");
+        todo_wine ok(ret, "getting DEVPKEY_Device_DriverDate failed with error %lu.\n", GetLastError());
+        todo_wine ok(proptype == DEVPROP_TYPE_FILETIME, "got unexpected proptype %#lx.\n", proptype);
+        todo_wine ok(len == sizeof(LARGE_INTEGER), "got unexpected length.\n");
+        todo_wine ok((filetime.QuadPart % 864000000000) == 0, "returned value should not contain a time.\n");
 
         ZeroMemory(buf, sizeof(buf));
         len = 0;
         proptype = DEVPROP_TYPE_EMPTY;
         ret = SetupDiGetDevicePropertyW(set, &devinfodata, &DEVPKEY_Device_DriverVersion,  &proptype, (BYTE *)buf, sizeof(buf), &len, 0);
-        ok(ret, "getting DEVPKEY_Device_DriverVersion failed with error %lu.\n", GetLastError());
-        ok(proptype == DEVPROP_TYPE_STRING, "got unexpected proptype %#lx.\n", proptype);
-        ok(len > 0, "got unexpected length.\n");
+        todo_wine ok(ret, "getting DEVPKEY_Device_DriverVersion failed with error %lu.\n", GetLastError());
+        todo_wine ok(proptype == DEVPROP_TYPE_STRING, "got unexpected proptype %#lx.\n", proptype);
+        todo_wine ok(len > 0, "got unexpected length.\n");
 
         ZeroMemory(buf, sizeof(buf));
         len = 0;
         proptype = DEVPROP_TYPE_EMPTY;
         ret = SetupDiGetDevicePropertyW(set, &devinfodata, &DEVPKEY_Device_DriverDesc,  &proptype, (BYTE *)buf, sizeof(buf), &len, 0);
-        ok(ret, "getting DEVPKEY_Device_DriverDesc failed with error %lu.\n", GetLastError());
-        ok(proptype == DEVPROP_TYPE_STRING, "got unexpected proptype %#lx.\n", proptype);
-        ok(len > 0, "got unexpected length.\n");
+        todo_wine ok(ret, "getting DEVPKEY_Device_DriverDesc failed with error %lu.\n", GetLastError());
+        todo_wine ok(proptype == DEVPROP_TYPE_STRING, "got unexpected proptype %#lx.\n", proptype);
+        todo_wine ok(len > 0, "got unexpected length.\n");
 
         ZeroMemory(buf, sizeof(buf));
         len = 0;
         proptype = DEVPROP_TYPE_EMPTY;
         ret = SetupDiGetDevicePropertyW(set, &devinfodata, &DEVPKEY_Device_DriverProvider,  &proptype, (BYTE *)buf, sizeof(buf), &len, 0);
-        ok(ret, "getting DEVPKEY_Device_DriverProvider failed with error %lu.\n", GetLastError());
-        ok(proptype == DEVPROP_TYPE_STRING, "got unexpected proptype %#lx.\n", proptype);
-        ok(len > 0, "got unexpected length.\n");
+        todo_wine ok(ret, "getting DEVPKEY_Device_DriverProvider failed with error %lu.\n", GetLastError());
+        todo_wine ok(proptype == DEVPROP_TYPE_STRING, "got unexpected proptype %#lx.\n", proptype);
+        todo_wine ok(len > 0, "got unexpected length.\n");
 
         ZeroMemory(buf, sizeof(buf));
         len = 0;
