diff --git a/dlls/msi/registry.c b/dlls/msi/registry.c
--- a/dlls/msi/registry.c
+++ b/dlls/msi/registry.c
@@ -452,6 +452,21 @@
     lstrcpyW( keypath, L"Software\\Microsoft\\Installer\\Features\\" );
     lstrcatW( keypath, squashed_pc );
     return RegDeleteTreeW(HKEY_CURRENT_USER, keypath);
+}
+
+static UINT MSIREG_OpenInstallerFeaturesKey(LPCWSTR szProduct, HKEY *key, BOOL create)
+{
+    REGSAM access = KEY_WOW64_64KEY | KEY_ALL_ACCESS;
+    WCHAR squashed_pc[SQUASHED_GUID_SIZE], keypath[0x200];
+
+    if (!squash_guid( szProduct, squashed_pc )) return ERROR_FUNCTION_FAILED;
+    TRACE("%s squashed %s\n", debugstr_w(szProduct), debugstr_w(squashed_pc));
+
+    lstrcpyW(keypath, L"Software\\Microsoft\\Windows\\CurrentVersion\\Installer\\Features\\");
+    lstrcatW( keypath, squashed_pc );
+
+    if (create) return RegCreateKeyExW(HKEY_LOCAL_MACHINE, keypath, 0, NULL, 0, access, NULL, key, NULL);
+    return RegOpenKeyExW(HKEY_LOCAL_MACHINE, keypath, 0, access, key);
 }
 
 UINT MSIREG_OpenUserDataFeaturesKey(LPCWSTR szProduct, LPCWSTR szUserSid, MSIINSTALLCONTEXT context,
@@ -953,16 +968,8 @@
     }
     if (szFeature)
     {
-        if (!len) /* Single feature was not encoded, figure it out */
-        {
-            if (MsiEnumFeaturesW( szProduct, 0, szFeature, NULL ) != ERROR_SUCCESS)
-                return ERROR_INVALID_PARAMETER;
-        }
-        else
-        {
-            memcpy( szFeature, &szDescriptor[20], len*sizeof(WCHAR) );
-            szFeature[len] = 0;
-        }
+        memcpy( szFeature, &szDescriptor[20], len*sizeof(WCHAR) );
+        szFeature[len] = 0;
     }
 
     len = p - szDescriptor + 1;
@@ -1072,18 +1079,13 @@
 {
     HKEY hkeyProduct = 0;
     DWORD r, sz;
-    MSIINSTALLCONTEXT context;
 
     TRACE( "%s, %lu, %p, %p\n", debugstr_w(szProduct), index, szFeature, szParent );
 
     if( !szProduct )
         return ERROR_INVALID_PARAMETER;
 
-    r = msi_locate_product(szProduct, &context);
-    if ( r != ERROR_SUCCESS )
-        return r;
-
-    r = MSIREG_OpenUserDataFeaturesKey(szProduct, NULL, context, &hkeyProduct, FALSE);
+    r = MSIREG_OpenInstallerFeaturesKey(szProduct,&hkeyProduct,FALSE);
     if( r != ERROR_SUCCESS )
         return ERROR_NO_MORE_ITEMS;
 
