diff --git a/dlls/gdi32/uniscribe/breaking.c b/dlls/gdi32/uniscribe/breaking.c
--- a/dlls/gdi32/uniscribe/breaking.c
+++ b/dlls/gdi32/uniscribe/breaking.c
@@ -32,6 +32,7 @@
 #include "winternl.h"
 
 #include "wine/debug.h"
+#include "wine/heap.h"
 #include "usp10_internal.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(uniscribe);
@@ -41,7 +42,7 @@
 enum breaking_types {
     b_BK=1, b_CR, b_LF, b_CM, b_SG, b_GL, b_CB, b_SP, b_ZW, b_NL, b_WJ, b_JL, b_JV, b_JT, b_H2, b_H3, b_XX, b_OP, b_CL,
     b_CP, b_QU, b_NS, b_EX, b_SY, b_IS, b_PR, b_PO, b_NU, b_AL, b_ID, b_IN, b_HY, b_BB, b_BA, b_SA, b_AI, b_B2, b_HL,
-    b_CJ, b_RI, b_EB, b_EM, b_ZWJ, b_AK, b_AP, b_AS, b_VF, b_VI, b_HH,
+    b_CJ, b_RI, b_EB, b_EM, b_ZWJ
 };
 
 enum breaking_class {b_r=1, b_s, b_x};
@@ -81,8 +82,8 @@
 
     TRACE("In      %s\n",debugstr_wn(chars,count));
 
-    break_class = malloc(count * sizeof(*break_class));
-    break_before = malloc(count * sizeof(*break_before));
+    break_class = heap_alloc(count * sizeof(*break_class));
+    break_before = heap_alloc(count * sizeof(*break_before));
 
     for (i = 0; i < count; i++)
     {
@@ -203,8 +204,7 @@
                 {
                     if (break_class[i-1] != b_SP &&
                         break_class[i-1] != b_BA &&
-                        break_class[i-1] != b_HY &&
-                        break_class[i-1] != b_HH)
+                        break_class[i-1] != b_HY)
                         else_break(&break_before[i],b_x);
                 }
                 break;
@@ -287,7 +287,6 @@
             /* LB21 */
             case b_BA:
             case b_HY:
-            case b_HH:
             case b_NS:
                 else_break(&break_before[i],b_x);
                 break;
@@ -301,7 +300,6 @@
                     switch (break_class[i+1])
                     {
                     case b_HY:
-                    case b_HH:
                     case b_BA:
                         else_break(&break_before[i+2], b_x);
                     }
@@ -441,6 +439,6 @@
         }
     }
 
-    free(break_before);
-    free(break_class);
+    heap_free(break_before);
+    heap_free(break_class);
 }
