diff --git a/dlls/ieframe/oleobject.c b/dlls/ieframe/oleobject.c
--- a/dlls/ieframe/oleobject.c
+++ b/dlls/ieframe/oleobject.c
@@ -69,11 +69,13 @@
 {
     WebBrowser *This;
 
+    static const WCHAR wszTHIS[] = {'T','H','I','S',0};
+
     if(msg == WM_CREATE) {
         This = *(WebBrowser**)lParam;
-        SetPropW(hwnd, L"THIS", This);
+        SetPropW(hwnd, wszTHIS, This);
     }else {
-        This = GetPropW(hwnd, L"THIS");
+        This = GetPropW(hwnd, wszTHIS);
     }
 
     switch(msg) {
@@ -98,6 +100,9 @@
     HWND parent = NULL;
     HRESULT hres;
 
+    static const WCHAR wszShellEmbedding[] =
+        {'S','h','e','l','l',' ','E','m','b','e','d','d','i','n','g',0};
+
     if(!shell_embedding_atom) {
         static WNDCLASSEXW wndclass = {
             sizeof(wndclass),
@@ -105,7 +110,7 @@
             shell_embedding_proc,
             0, 0 /* native uses 8 */, NULL, NULL, NULL,
             (HBRUSH)(COLOR_WINDOW + 1), NULL,
-            L"Shell Embedding",
+            wszShellEmbedding,
             NULL
         };
         wndclass.hInstance = ieframe_instance;
@@ -121,7 +126,7 @@
 
     This->doc_host.frame_hwnd = This->shell_embedding_hwnd = CreateWindowExW(
             WS_EX_WINDOWEDGE,
-            L"Shell Embedding", L"Shell Embedding",
+            wszShellEmbedding, wszShellEmbedding,
             WS_CLIPSIBLINGS | WS_CLIPCHILDREN
             | (parent ? WS_CHILD | WS_TABSTOP : WS_POPUP | WS_MAXIMIZEBOX),
             0, 0, 0, 0, parent,
@@ -196,6 +201,8 @@
 {
     HRESULT hres;
 
+    static const WCHAR wszitem[] = {'i','t','e','m',0};
+
     if(This->inplace)
     {
         if(This->shell_embedding_hwnd)
@@ -213,9 +220,9 @@
     IOleInPlaceSiteEx_OnUIActivate(This->inplace);
 
     if(This->doc_host.frame)
-        IOleInPlaceFrame_SetActiveObject(This->doc_host.frame, &This->IOleInPlaceActiveObject_iface, L"item");
+        IOleInPlaceFrame_SetActiveObject(This->doc_host.frame, &This->IOleInPlaceActiveObject_iface, wszitem);
     if(This->uiwindow)
-        IOleInPlaceUIWindow_SetActiveObject(This->uiwindow, &This->IOleInPlaceActiveObject_iface, L"item");
+        IOleInPlaceUIWindow_SetActiveObject(This->uiwindow, &This->IOleInPlaceActiveObject_iface, wszitem);
 
     if(This->doc_host.frame)
         IOleInPlaceFrame_SetMenu(This->doc_host.frame, NULL, NULL, This->shell_embedding_hwnd);
