diff --git a/dlls/wintab32/context.c b/dlls/wintab32/context.c
--- a/dlls/wintab32/context.c
+++ b/dlls/wintab32/context.c
@@ -26,7 +26,7 @@
 #include "windef.h"
 #include "winerror.h"
 #include "winbase.h"
-#include "ntuser.h"
+#include "winuser.h"
 #include "winnls.h"
 
 #include "wintab.h"
@@ -149,7 +149,7 @@
 
     if (loaded == TI_START)
     {
-        if (NtUserMessageCall(hwndDefault, NtUserWintabInit, 0, 0, NULL, NtUserWintabDriverCall, FALSE))
+        if (pLoadTabletInfo && pLoadTabletInfo(hwndDefault))
         {
             TRACE("Initialized the tablet to hwnd %p\n", hwndDefault);
             loaded = TI_OK;
@@ -373,10 +373,6 @@
     memset(lpPkt,0,rc);
 }
 
-static UINT DriverInfo(UINT category, UINT index, void *output)
-{
-    return NtUserMessageCall(NULL, NtUserWintabInfo, category, index, output, NtUserWintabDriverCall, FALSE);
-}
 
 static UINT WTInfoT(UINT wCategory, UINT nIndex, LPVOID lpOutput, BOOL bUnicode)
 {
@@ -410,7 +406,7 @@
         if (lpOutput)
         {
             LOGCONTEXTW buf;
-            DriverInfo(wCategory, nIndex, &buf);
+            pWTInfoW(wCategory, nIndex, &buf);
 
             /*  Handle system extents here, as we can use user32.dll code to set them */
             if(wCategory == WTI_DEFSYSCTX)
@@ -429,14 +425,14 @@
     }
     else if (is_string_field(wCategory, nIndex) && !bUnicode)
     {
-        int size = DriverInfo(wCategory, nIndex, NULL);
+        int size = pWTInfoW(wCategory, nIndex, NULL);
         WCHAR *buf = malloc(size);
-        DriverInfo(wCategory, nIndex, buf);
+        pWTInfoW(wCategory, nIndex, buf);
         result = WideCharToMultiByte(CP_ACP, 0, buf, size/sizeof(WCHAR), lpOutput, lpOutput ? 2*size : 0, NULL, NULL);
         free(buf);
     }
     else
-        result =  DriverInfo(wCategory, nIndex, lpOutput);
+        result =  pWTInfoW(wCategory, nIndex, lpOutput);
 
     TRACE("returns %d\n", result);
     return result;
@@ -485,7 +481,7 @@
     gOpenContexts = newcontext;
     LeaveCriticalSection(&csTablet);
 
-    NtUserMessageCall(hWnd, NtUserWintabAttach, 0, 0, NULL, NtUserWintabDriverCall, FALSE);
+    pAttachEventQueueToTablet(hWnd);
 
     TABLET_PostTabletMessage(newcontext, _WT_CTXOPEN(newcontext->context.lcMsgBase), (WPARAM)newcontext->handle,
                       newcontext->context.lcStatus, TRUE);
