diff --git a/dlls/gdi32/dc.c b/dlls/gdi32/dc.c
--- a/dlls/gdi32/dc.c
+++ b/dlls/gdi32/dc.c
@@ -1142,8 +1142,7 @@
  */
 BOOL WINAPI FixBrushOrgEx( HDC hdc, INT x, INT y, POINT *oldorg )
 {
-    /* From Windows 2000 this is a NOP returning FALSE */
-    return FALSE;
+    return SetBrushOrgEx( hdc, x, y, oldorg );
 }
 
 /***********************************************************************
@@ -1440,6 +1439,17 @@
     return ret;
 }
 
+/***********************************************************************
+ *           GetMiterLimit  (GDI32.@)
+ */
+BOOL WINAPI GetMiterLimit( HDC hdc, FLOAT *limit )
+{
+    DC_ATTR *dc_attr;
+    if (!(dc_attr = get_dc_attr( hdc ))) return FALSE;
+    if (limit) *limit = dc_attr->miter_limit;
+    return TRUE;
+}
+
 /*******************************************************************
  *           SetMiterLimit  (GDI32.@)
  */
@@ -1449,7 +1459,9 @@
     if (!(dc_attr = get_dc_attr( hdc ))) return FALSE;
     if (dc_attr->emf && !EMFDC_SetMiterLimit( dc_attr, limit )) return 0;
     if (limit < 1.0f) return FALSE;
-    return NtGdiSetMiterLimit( hdc, *(DWORD *)&limit, old_limit );
+    if (old_limit) *old_limit = dc_attr->miter_limit;
+    dc_attr->miter_limit = limit;
+    return TRUE;
 }
 
 /***********************************************************************
@@ -2377,10 +2389,15 @@
     return NtGdiScaleWindowExtEx( hdc, x_num, x_denom, y_num, y_denom, size );
 }
 
+static UINT WINAPI realize_palette( HDC hdc )
+{
+    return NtUserRealizePalette( hdc );
+}
+
 /* Pointers to USER implementation of SelectPalette/RealizePalette */
 /* they will be patched by USER on startup */
 HPALETTE (WINAPI *pfnSelectPalette)( HDC hdc, HPALETTE hpal, WORD bkgnd ) = NtUserSelectPalette;
-UINT (WINAPI *pfnRealizePalette)( HDC hdc ) = NtUserRealizePalette;
+UINT (WINAPI *pfnRealizePalette)( HDC hdc ) = realize_palette;
 
 /***********************************************************************
  *           SelectPalette    (GDI32.@)
@@ -2423,7 +2440,8 @@
  */
 BOOL WINAPI CancelDC(HDC hdc)
 {
-    return NtGdiCancelDC( hdc );
+    FIXME( "stub\n" );
+    return TRUE;
 }
 
 /***********************************************************************
