diff --git a/dlls/d3dcompiler_43/tests/hlsl_d3d9.c b/dlls/d3dcompiler_43/tests/hlsl_d3d9.c
--- a/dlls/d3dcompiler_43/tests/hlsl_d3d9.c
+++ b/dlls/d3dcompiler_43/tests/hlsl_d3d9.c
@@ -561,6 +561,7 @@
         return;
     device = test_context.device;
 
+    todo_wine
     ps_code = compile_shader(ps_if_source, "ps_2_0", 0);
     if (ps_code)
     {
@@ -570,15 +571,15 @@
         for (i = 0; i < 200; i += 40)
         {
             v = get_readback_vec4(&rb, i, 0);
-            ok(compare_vec4(v, 0.9f, 0.8f, 0.7f, 0.6f, 0),
-                    "Got unexpected value {%.8e, %.8e, %.8e, %.8e}.\n", v->x, v->y, v->z, v->w);
+            todo_wine ok(compare_vec4(v, 0.9f, 0.8f, 0.7f, 0.6f, 0),
+                         "Got unexpected value {%.8e, %.8e, %.8e, %.8e}.\n", v->x, v->y, v->z, v->w);
         }
 
         for (i = 240; i < 640; i += 40)
         {
             v = get_readback_vec4(&rb, i, 0);
-            ok(compare_vec4(v, 0.1f, 0.2f, 0.3f, 0.4f, 0),
-                    "Got unexpected value {%.8e, %.8e, %.8e, %.8e}.\n", v->x, v->y, v->z, v->w);
+            todo_wine ok(compare_vec4(v, 0.1f, 0.2f, 0.3f, 0.4f, 0),
+                         "Got unexpected value {%.8e, %.8e, %.8e, %.8e}.\n", v->x, v->y, v->z, v->w);
         }
 
         release_readback(&rb);
@@ -706,9 +707,9 @@
     init_readback(device, &rb);
     for (i = 0; i < 32; ++i)
     {
-        float expect_x = (sinf((i + 0.5f) * 2.0f * M_PI / 32) + 1.0f) / 2.0f;
-        float expect_y = (cosf((i + 0.5f) * 2.0f * M_PI / 32) + 1.0f) / 2.0f;
-        v = get_readback_vec4(&rb, i * 640 / 32 + 640 / 2 / 32, 0);
+        float expect_x = (sinf(i * 2 * M_PI / 32) + 1.0f) / 2.0f;
+        float expect_y = (cosf(i * 2 * M_PI / 32) + 1.0f) / 2.0f;
+        v = get_readback_vec4(&rb, i * 640 / 32, 0);
         ok(compare_vec4(v, expect_x, expect_y, 0.0f, 0.0f, 4096),
                 "Test %u: Got {%.8e, %.8e, %.8e, %.8e}, expected {%.8e, %.8e, %.8e, %.8e}.\n",
                 i, v->x, v->y, v->z, v->w, expect_x, expect_y, 0.0f, 0.0f);
@@ -1192,7 +1193,6 @@
     ok(desc->Elements == expect->Elements, "%s: got Elements %u.\n", prefix, desc->Elements);
     ok(desc->StructMembers == expect->StructMembers, "%s: got StructMembers %u.\n", prefix, desc->StructMembers);
     ok(desc->Bytes == expect->Bytes, "%s: got Bytes %u.\n", prefix, desc->Bytes);
-    todo_wine_if(nonzero_defaultvalue)
     ok(!!desc->DefaultValue == nonzero_defaultvalue, "%s: got DefaultValue %p.\n", prefix, desc->DefaultValue);
 }
 
@@ -1287,7 +1287,9 @@
     static const D3DXCONSTANT_DESC expect_fields_j =
         {"a", D3DXRS_FLOAT4, 0, 3, D3DXPC_MATRIX_COLUMNS, D3DXPT_FLOAT, 3, 3, 1, 0, 36};
 
-    ps_code = compile_shader(source, "ps_2_0", 0);
+    todo_wine ps_code = compile_shader(source, "ps_2_0", 0);
+    if (!ps_code)
+        return;
 
     hr = pD3DXGetShaderConstantTable(ID3D10Blob_GetBufferPointer(ps_code), &constants);
     ok(hr == D3D_OK, "Got unexpected hr %#lx.\n", hr);
@@ -1787,7 +1789,6 @@
     delete_file(L"include\\include1.h");
     delete_file(L"include1.h");
     delete_file(L"include\\include2.h");
-    delete_file(L"include\\include3.h");
     delete_directory(L"include");
 }
 
