diff --git a/dlls/msxml3/pi.c b/dlls/msxml3/pi.c
--- a/dlls/msxml3/pi.c
+++ b/dlls/msxml3/pi.c
@@ -191,7 +191,8 @@
     hr = IXMLDOMProcessingInstruction_get_nodeName(iface, &target);
     if(hr == S_OK)
     {
-        if(!wcscmp(target, L"xml"))
+        static const WCHAR xmlW[] = {'x','m','l',0};
+        if(!wcscmp(target, xmlW))
         {
             SysFreeString(target);
             return E_FAIL;
@@ -419,6 +420,7 @@
     IXMLDOMNamedNodeMap** map)
 {
     dom_pi *This = impl_from_IXMLDOMProcessingInstruction( iface );
+    static const WCHAR xmlW[] = {'x','m','l',0};
     HRESULT hr;
     BSTR name;
 
@@ -431,7 +433,7 @@
     hr = node_get_nodeName(&This->node, &name);
     if (hr != S_OK) return hr;
 
-    if (!wcscmp(name, L"xml"))
+    if (!wcscmp(name, xmlW))
         *map = create_nodemap(This->node.node, &dom_pi_attr_map);
 
     SysFreeString(name);
@@ -514,10 +516,12 @@
     BSTR* p)
 {
     dom_pi *This = impl_from_IXMLDOMProcessingInstruction( iface );
+    static const WCHAR processinginstructionW[] =
+        {'p','r','o','c','e','s','s','i','n','g','i','n','s','t','r','u','c','t','i','o','n',0};
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return return_bstr(L"processinginstruction", p);
+    return return_bstr(processinginstructionW, p);
 }
 
 static HRESULT WINAPI dom_pi_get_text(
@@ -730,7 +734,8 @@
     hr = IXMLDOMProcessingInstruction_get_nodeName(iface, &target);
     if(hr == S_OK)
     {
-        if(!wcscmp(target, L"xml"))
+        static const WCHAR xmlW[] = {'x','m','l',0};
+        if(!wcscmp(target, xmlW))
         {
             SysFreeString(target);
             return E_FAIL;
@@ -744,6 +749,7 @@
 
 HRESULT dom_pi_put_xml_decl(IXMLDOMNode *node, BSTR data)
 {
+    static const WCHAR xmlW[] = {'x','m','l',0};
     xmlnode *node_obj;
     HRESULT hr;
     BSTR name;
@@ -760,7 +766,7 @@
     if (FAILED(hr))
         return hr;
 
-    if (!wcscmp(name, L"xml") && !node_obj->node->properties)
+    if (!lstrcmpW(name, xmlW) && !node_obj->node->properties)
         hr = parse_xml_decl(node_obj->node);
     else
         hr = S_OK;
