diff --git a/dlls/msvfw32/mciwnd.c b/dlls/msvfw32/mciwnd.c
--- a/dlls/msvfw32/mciwnd.c
+++ b/dlls/msvfw32/mciwnd.c
@@ -551,15 +551,16 @@
     case MCIWNDM_OPENW:
         {
             RECT rc;
+            BOOL has_window;
             HCURSOR hCursor;
             MCI_OPEN_PARMSW mci_open;
             MCI_GETDEVCAPS_PARMS mci_devcaps;
+            MCI_DGV_WINDOW_PARMSW mci_window;
             WCHAR aliasW[64];
             WCHAR drv_name[MAX_PATH];
             static const WCHAR formatW[] = {'%','d',0};
             static const WCHAR mci32W[] = {'m','c','i','3','2',0};
             static const WCHAR system_iniW[] = {'s','y','s','t','e','m','.','i','n','i',0};
-            static const WCHAR avivideoW[] = {'a','v','i','v','i','d','e','o',0};
 
             TRACE("MCIWNDM_OPENW %s\n", debugstr_w((LPWSTR)lParam));
 
@@ -584,16 +585,6 @@
             mwi->lasterror = mciSendCommandW(mwi->mci, MCI_OPEN,
                                              MCI_OPEN_ELEMENT | MCI_OPEN_ALIAS | MCI_WAIT,
                                              (DWORD_PTR)&mci_open);
-
-            if (mwi->lasterror)
-            {
-                mci_open.lpstrDeviceType = avivideoW;
-                mwi->lasterror = mciSendCommandW(mwi->mci, MCI_OPEN,
-                                                 MCI_OPEN_ELEMENT | MCI_OPEN_TYPE |
-                                                 MCI_OPEN_ALIAS | MCI_WAIT,
-                                                 (DWORD_PTR)&mci_open);
-            }
-
             SetCursor(hCursor);
 
             if (mwi->lasterror && !(mwi->dwStyle & MCIWNDF_NOERRORDLG))
@@ -626,16 +617,29 @@
 
             mwi->dev_type = mci_devcaps.dwReturn;
 
+            mci_window.nCmdShow = -1;
+            mwi->lasterror = mciSendCommandW(mwi->mci, MCI_WINDOW, MCI_DGV_WINDOW_STATE, (DWORD_PTR)&mci_window);
+
+            if (mwi->lasterror == MCIERR_NO_WINDOW)
+            {
+                has_window = 0;
+                mwi->lasterror = 0;
+            }
+            else if (mwi->lasterror)
+            {
+                MCIWND_notify_error(mwi);
+                goto end_of_mci_open;
+            }
+            else has_window = 1;
+
             drv_name[0] = 0;
             SendMessageW(hWnd, MCIWNDM_GETDEVICEW, 256, (LPARAM)drv_name);
             if (drv_name[0] && GetPrivateProfileStringW(mci32W, drv_name, NULL,
                                             drv_name, MAX_PATH, system_iniW))
                 mwi->hdrv = OpenDriver(drv_name, NULL, 0);
 
-            if (mwi->dev_type == MCI_DEVTYPE_DIGITAL_VIDEO)
-            {
-                MCI_DGV_WINDOW_PARMSW mci_window;
-
+            if (mwi->dev_type == MCI_DEVTYPE_DIGITAL_VIDEO && has_window)
+            {
                 mci_window.hWnd = hWnd;
                 mwi->lasterror = mciSendCommandW(mwi->mci, MCI_WINDOW,
                                                  MCI_DGV_WINDOW_HWND,
