diff --git a/dlls/joy.cpl/main.c b/dlls/joy.cpl/main.c
--- a/dlls/joy.cpl/main.c
+++ b/dlls/joy.cpl/main.c
@@ -189,7 +189,7 @@
 
 static DWORD get_advanced_option( const WCHAR *option, DWORD default_value )
 {
-    DWORD value, size = sizeof(value);
+    DWORD value, size;
     HKEY hkey;
     if (!get_advanced_key( &hkey )) return default_value;
     if (RegGetValueW( hkey, NULL, option, RRF_RT_REG_DWORD, NULL, (BYTE *)&value, &size ))
@@ -449,12 +449,6 @@
             .pszTemplate = MAKEINTRESOURCEW( IDD_TEST_XI ),
             .pfnDlgProc = test_xi_dialog_proc,
         },
-        {
-            .dwSize = sizeof(PROPSHEETPAGEW),
-            .hInstance = hcpl,
-            .pszTemplate = MAKEINTRESOURCEW( IDD_TEST_WGI ),
-            .pfnDlgProc = test_wgi_dialog_proc,
-        },
     };
     PROPSHEETHEADERW header =
     {
@@ -518,47 +512,15 @@
         .lpfnWndProc = &test_di_buttons_window_proc,
         .lpszClassName = L"JoyCplDInputButtons",
     };
-    WNDCLASSW wgi_axes_class =
-    {
-        .hInstance = hcpl,
-        .lpfnWndProc = &test_wgi_axes_window_proc,
-        .lpszClassName = L"JoyCplWGIAxes",
-    };
-    WNDCLASSW wgi_povs_class =
-    {
-        .hInstance = hcpl,
-        .lpfnWndProc = &test_wgi_povs_window_proc,
-        .lpszClassName = L"JoyCplWGIPOVs",
-    };
-    WNDCLASSW wgi_buttons_class =
-    {
-        .hInstance = hcpl,
-        .lpfnWndProc = &test_wgi_buttons_window_proc,
-        .lpszClassName = L"JoyCplWGIButtons",
-    };
-    WNDCLASSW wgi_gamepad_class =
-    {
-        .hInstance = hcpl,
-        .lpfnWndProc = &test_wgi_gamepad_window_proc,
-        .lpszClassName = L"JoyCplWGIGamepad",
-    };
 
     RegisterClassW( &xi_class );
     RegisterClassW( &di_axes_class );
     RegisterClassW( &di_povs_class );
     RegisterClassW( &di_buttons_class );
-    RegisterClassW( &wgi_axes_class );
-    RegisterClassW( &wgi_povs_class );
-    RegisterClassW( &wgi_buttons_class );
-    RegisterClassW( &wgi_gamepad_class );
 }
 
 static void unregister_window_class(void)
 {
-    UnregisterClassW( L"JoyCplWGIAxes", hcpl );
-    UnregisterClassW( L"JoyCplWGIPovs", hcpl );
-    UnregisterClassW( L"JoyCplWGIButtons", hcpl );
-    UnregisterClassW( L"JoyCplWGIGamepad", hcpl );
     UnregisterClassW( L"JoyCplDInputAxes", hcpl );
     UnregisterClassW( L"JoyCplDInputPOVs", hcpl );
     UnregisterClassW( L"JoyCplDInputButtons", hcpl );
