diff --git a/dlls/dwrite/freetype.c b/dlls/dwrite/freetype.c
--- a/dlls/dwrite/freetype.c
+++ b/dlls/dwrite/freetype.c
@@ -44,7 +44,7 @@
 
 #include "dwrite_private.h"
 
-#ifdef SONAME_LIBFREETYPE
+#ifdef HAVE_FREETYPE
 
 WINE_DEFAULT_DEBUG_CHANNEL(dwrite);
 
@@ -80,13 +80,14 @@
 MAKE_FUNCPTR(FT_Outline_Copy);
 MAKE_FUNCPTR(FT_Outline_Decompose);
 MAKE_FUNCPTR(FT_Outline_Done);
-MAKE_FUNCPTR(FT_Outline_EmboldenXY);
+MAKE_FUNCPTR(FT_Outline_Embolden);
 MAKE_FUNCPTR(FT_Outline_Get_Bitmap);
 MAKE_FUNCPTR(FT_Outline_New);
 MAKE_FUNCPTR(FT_Outline_Transform);
 MAKE_FUNCPTR(FT_Outline_Translate);
 MAKE_FUNCPTR(FT_Set_Pixel_Sizes);
 #undef MAKE_FUNCPTR
+static FT_Error (*pFT_Outline_EmboldenXY)(FT_Outline *, FT_Pos, FT_Pos);
 
 #define FaceFromObject(o) ((FT_Face)(ULONG_PTR)(o))
 
@@ -146,13 +147,14 @@
     LOAD_FUNCPTR(FT_Outline_Copy)
     LOAD_FUNCPTR(FT_Outline_Decompose)
     LOAD_FUNCPTR(FT_Outline_Done)
-    LOAD_FUNCPTR(FT_Outline_EmboldenXY)
+    LOAD_FUNCPTR(FT_Outline_Embolden)
     LOAD_FUNCPTR(FT_Outline_Get_Bitmap)
     LOAD_FUNCPTR(FT_Outline_New)
     LOAD_FUNCPTR(FT_Outline_Transform)
     LOAD_FUNCPTR(FT_Outline_Translate)
     LOAD_FUNCPTR(FT_Set_Pixel_Sizes)
 #undef LOAD_FUNCPTR
+    pFT_Outline_EmboldenXY = dlsym(ft_handle, "FT_Outline_EmboldenXY");
 
     if (pFT_Init_FreeType(&library) != 0)
     {
@@ -427,7 +429,10 @@
     FT_Pos strength;
 
     strength = pFT_MulDiv(emsize, 1 << 6, 24);
-    pFT_Outline_EmboldenXY(outline, strength, 0);
+    if (pFT_Outline_EmboldenXY)
+        pFT_Outline_EmboldenXY(outline, strength, 0);
+    else
+        pFT_Outline_Embolden(outline, strength);
 }
 
 static void embolden_glyph(FT_Glyph glyph, FLOAT emsize)
@@ -735,7 +740,7 @@
     return STATUS_SUCCESS;
 }
 
-#else /* SONAME_LIBFREETYPE */
+#else /* HAVE_FREETYPE */
 
 static NTSTATUS process_attach(void *args)
 {
@@ -794,7 +799,7 @@
     return STATUS_NOT_IMPLEMENTED;
 }
 
-#endif /* SONAME_LIBFREETYPE */
+#endif /* HAVE_FREETYPE */
 
 const unixlib_entry_t __wine_unix_call_funcs[] =
 {
