diff --git a/dlls/krnl386.exe16/global.c b/dlls/krnl386.exe16/global.c
--- a/dlls/krnl386.exe16/global.c
+++ b/dlls/krnl386.exe16/global.c
@@ -104,14 +104,14 @@
  * Create a global heap block for a fixed range of linear memory.
  */
 HGLOBAL16 GLOBAL_CreateBlock( WORD flags, void *ptr, DWORD size,
-                              HGLOBAL16 hOwner, struct ldt_bits bits )
+                              HGLOBAL16 hOwner, unsigned char selflags )
 {
     WORD sel, selcount;
     GLOBALARENA *pArena;
 
       /* Allocate the selector(s) */
 
-    sel = SELECTOR_AllocBlock( ptr, size, bits );
+    sel = SELECTOR_AllocBlock( ptr, size, selflags );
     if (!sel) return 0;
     selcount = (size + 0xffff) / 0x10000;
 
@@ -132,7 +132,7 @@
     pArena->flags = flags & GA_MOVEABLE;
     if (flags & GMEM_DISCARDABLE) pArena->flags |= GA_DISCARDABLE;
     if (flags & GMEM_DDESHARE) pArena->flags |= GA_IPCSHARE;
-    if (!(bits.type & 0x08 /* code */)) pArena->flags |= GA_DGROUP;
+    if (!(selflags & (LDT_FLAGS_CODE ^ LDT_FLAGS_DATA))) pArena->flags |= GA_DGROUP;
     pArena->selCount = selcount;
     if (selcount > 1)  /* clear the next arena blocks */
         memset( pArena + 1, 0, (selcount - 1) * sizeof(GLOBALARENA) );
@@ -192,7 +192,7 @@
  *
  * Implementation of GlobalAlloc16()
  */
-HGLOBAL16 GLOBAL_Alloc( UINT16 flags, DWORD size, HGLOBAL16 hOwner, struct ldt_bits bits )
+HGLOBAL16 GLOBAL_Alloc( UINT16 flags, DWORD size, HGLOBAL16 hOwner, unsigned char selflags )
 {
     void *ptr;
     HGLOBAL16 handle;
@@ -202,7 +202,7 @@
 
     /* If size is 0, create a discarded block */
 
-    if (size == 0) return GLOBAL_CreateBlock( flags, NULL, 1, hOwner, bits );
+    if (size == 0) return GLOBAL_CreateBlock( flags, NULL, 1, hOwner, selflags );
 
     /* Fixup the size */
 
@@ -217,7 +217,7 @@
 
       /* Allocate the selector(s) */
 
-    handle = GLOBAL_CreateBlock( flags, ptr, size, hOwner, bits );
+    handle = GLOBAL_CreateBlock( flags, ptr, size, hOwner, selflags );
     if (!handle)
     {
         HeapFree( get_win16_heap(), 0, ptr );
@@ -250,7 +250,7 @@
         STACK16FRAME *frame = CURRENT_STACK16;
         owner = GetExePtr( frame->cs );
     }
-    return GLOBAL_Alloc( flags, size, owner, data_segment );
+    return GLOBAL_Alloc( flags, size, owner, LDT_FLAGS_DATA );
 }
 
 
@@ -775,7 +775,7 @@
        WORD	 wSelector;
        GLOBALARENA *pArena;
 
-       wSelector = GLOBAL_CreateBlock(GMEM_FIXED, lpBlock, size, hModule, data_segment );
+       wSelector = GLOBAL_CreateBlock(GMEM_FIXED, lpBlock, size, hModule, LDT_FLAGS_DATA );
        pArena = GET_ARENA_PTR(wSelector);
        pArena->flags |= GA_DOSMEM;
        return MAKELONG(wSelector,uParagraph);
