diff --git a/dlls/opengl32/unix_private.h b/dlls/opengl32/unix_private.h
--- a/dlls/opengl32/unix_private.h
+++ b/dlls/opengl32/unix_private.h
@@ -21,7 +21,6 @@
 #include <stdarg.h>
 #include <stddef.h>
 #include <stdlib.h>
-#include <pthread.h>
 
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
@@ -31,14 +30,13 @@
 #include "wingdi.h"
 #include "ntgdi.h"
 
-#include "wine/opengl_driver.h"
-#include "unix_thunks.h"
+#include "wine/wgl.h"
+#include "wine/wgl_driver.h"
 
 struct registry_entry
 {
     const char *name;      /* name of the extension */
     const char *extension; /* name of the GL/WGL extension */
-    size_t offset;         /* offset in the opengl_funcs table */
 };
 
 extern const struct registry_entry extension_registry[];
@@ -48,16 +46,11 @@
 
 static inline const struct opengl_funcs *get_dc_funcs( HDC hdc )
 {
-    DWORD has_opengl;
-
-    if (NtGdiGetDCDword( hdc, NtGdiHasOpenGL, &has_opengl ) && has_opengl)
-        return __wine_get_opengl_driver( WINE_OPENGL_DRIVER_VERSION );
-
-    RtlSetLastWin32Error( ERROR_INVALID_HANDLE );
-    return &null_opengl_funcs;
+    const struct opengl_funcs *funcs = __wine_get_wgl_driver( hdc, WINE_WGL_DRIVER_VERSION );
+    if (!funcs) RtlSetLastWin32Error( ERROR_INVALID_HANDLE );
+    else if (funcs == (void *)-1) funcs = &null_opengl_funcs;
+    return funcs;
 }
-
-#ifdef _WIN64
 
 static inline void *copy_wow64_ptr32s( UINT_PTR address, ULONG count )
 {
@@ -75,24 +68,4 @@
     return (TEB *)((char *)teb32_ptr + teb32_ptr->WowTebOffset);
 }
 
-extern struct buffer *invalidate_buffer_name( TEB *teb, GLuint name );
-extern struct buffer *invalidate_buffer_target( TEB *teb, GLenum target );
-extern void free_buffer( const struct opengl_funcs *funcs, struct buffer *buffer );
-extern NTSTATUS return_wow64_string( const void *str, PTR32 *wow64_str );
-
-#endif
-
-extern pthread_mutex_t wgl_lock;
-
-extern NTSTATUS process_attach( void *args );
-extern NTSTATUS thread_attach( void *args );
-extern NTSTATUS process_detach( void *args );
-extern NTSTATUS get_pixel_formats( void *args );
-extern void set_context_attribute( TEB *teb, GLenum name, const void *value, size_t size );
-extern void set_current_fbo( TEB *teb, GLenum target, GLuint framebuffer );
-extern GLuint get_default_fbo( TEB *teb, GLenum target );
-extern void push_default_fbo( TEB *teb );
-extern void pop_default_fbo( TEB *teb );
-extern void resolve_default_fbo( TEB *teb, BOOL read );
-
 #endif /* __WINE_OPENGL32_UNIX_PRIVATE_H */
