diff --git a/dlls/win87em.dll16/win87em.c b/dlls/win87em.dll16/win87em.c
--- a/dlls/win87em.dll16/win87em.c
+++ b/dlls/win87em.dll16/win87em.c
@@ -61,7 +61,11 @@
 {
     context->Eax &= ~0xffff;  /* set AX to 0 */
     if (Installed)
+#ifdef __i386__
         __asm__("fclex");
+#else
+        ;
+#endif
     StatusWord_3 = StatusWord_2 = 0;
 }
 
@@ -71,7 +75,9 @@
     context->Eax &= ~0x00c3;
     if (Installed) {
         CtrlWord_Internal = LOWORD(context->Eax);
+#ifdef __i386__
         __asm__("wait;fldcw %0" : : "m" (CtrlWord_Internal));
+#endif
     }
     CtrlWord_2 = LOWORD(context->Eax);
 }
@@ -79,8 +85,10 @@
 static void WIN87_Init( CONTEXT *context )
 {
     if (Installed) {
+#ifdef __i386__
         __asm__("fninit");
         __asm__("fninit");
+#endif
     }
     StackBottom = StackTop;
     context->Eax = (context->Eax & ~0xffff) | 0x1332;
@@ -89,9 +97,9 @@
 }
 
 /***********************************************************************
- *		__fpMath (WIN87EM.1)
- */
-void WINAPI __fpMath( CONTEXT *context )
+ *		_fpMath (WIN87EM.1)
+ */
+void WINAPI _fpMath( CONTEXT *context )
 {
     TRACE("(cs:eip=%04lx:%04lx es=%04lx bx=%04lx ax=%04lx dx=%04lx)\n",
           context->SegCs, context->Eip, context->SegEs, context->Ebx,
@@ -145,12 +153,14 @@
             /* I don't know much about asm() programming. This could be
              * wrong.
              */
+#ifdef __i386__
            __asm__ __volatile__("fstcw %0;wait" : "=m" (save) : : "memory");
            __asm__ __volatile__("fstcw %0;wait" : "=m" (mask) : : "memory");
            __asm__ __volatile__("orw $0xC00,%0" : "=m" (mask) : : "memory");
            __asm__ __volatile__("fldcw %0;wait" : : "m" (mask));
            __asm__ __volatile__("frndint");
            __asm__ __volatile__("fldcw %0" : : "m" (save));
+#endif
         }
         break;
 
@@ -173,7 +183,9 @@
     case 8: /* restore internal status words from emulator status word */
         context->Eax &= ~0xffff;  /* set AX to 0 */
         if (Installed) {
+#ifdef __i386__
             __asm__("fstsw %0;wait" : "=m" (StatusWord_1));
+#endif
             context->Eax |= StatusWord_1 & 0x3f;
         }
         context->Eax = (context->Eax | StatusWord_2) & ~0xe000;
