diff --git a/dlls/wpcap/tests/wpcap.c b/dlls/wpcap/tests/wpcap.c
--- a/dlls/wpcap/tests/wpcap.c
+++ b/dlls/wpcap/tests/wpcap.c
@@ -86,7 +86,7 @@
 static const char * (CDECL *ppcap_datalink_val_to_description)( int );
 static const char * (CDECL *ppcap_datalink_val_to_name)( int );
 static int (CDECL *ppcap_dispatch)( pcap_t *, int,
-                                    void (CDECL *)(unsigned char *, const struct pcap_pkthdr *, const unsigned char *),
+                                    void (CALLBACK *)(unsigned char *, const struct pcap_pkthdr *, const unsigned char *),
                                     unsigned char * );
 static pcap_dumper_t * (CDECL *ppcap_dump_open)( pcap_t *, const char * );
 static void (CDECL *ppcap_dump)( unsigned char *, const struct pcap_pkthdr *, const unsigned char * );
@@ -107,11 +107,10 @@
 static char * (CDECL *ppcap_lookupdev)( char * );
 static int (CDECL *ppcap_lookupnet)( const char *, unsigned int *, unsigned int *, char * );
 static int (CDECL *ppcap_loop)( pcap_t *, int,
-                                void (CDECL *)(unsigned char *, const struct pcap_pkthdr *, const unsigned char *),
+                                void (CALLBACK *)(unsigned char *, const struct pcap_pkthdr *, const unsigned char *),
                                 unsigned char * );
 static int (CDECL *ppcap_set_buffer_size)( pcap_t *, int );
 static int (CDECL *ppcap_set_datalink)( pcap_t *, int );
-static int (CDECL *ppcap_set_immediate_mode)( pcap_t *, int );
 static int (CDECL *ppcap_set_promisc)( pcap_t *, int );
 static int (CDECL *ppcap_set_timeout)( pcap_t *, int );
 static int (CDECL *ppcap_set_tstamp_precision)( pcap_t *, int );
@@ -122,9 +121,9 @@
 static const char * (CDECL *ppcap_tstamp_type_val_to_description)( int );
 static const char * (CDECL *ppcap_tstamp_type_val_to_name)( int );
 
-static void CDECL capture_callback( unsigned char *user, const struct pcap_pkthdr *hdr, const unsigned char *bytes )
-{
-    trace( "user %p hdr %p bytes %p\n", user, hdr, bytes );
+static void CALLBACK capture_callback( unsigned char *user, const struct pcap_pkthdr *hdr, const unsigned char *bytes )
+{
+    trace( "user %p hdr %p byte %p\n", user, hdr, bytes );
 }
 
 static void test_capture( void )
@@ -143,9 +142,6 @@
     pcap = ppcap_create( dev, errbuf );
     ok( pcap != NULL, "got NULL (%s)\n", errbuf );
 
-    ret = ppcap_set_immediate_mode( pcap, 1 );
-    ok( !ret, "got %d\n", ret );
-
     ret = ppcap_set_promisc( pcap, 1 );
     ok( !ret, "got %d\n", ret );
 
@@ -270,7 +266,7 @@
     ok( !strcmp(str, "Ethernet"), "got %s\n", wine_dbgstr_a(str) );
 }
 
-static void CDECL dump_callback( unsigned char *user, const struct pcap_pkthdr *hdr, const unsigned char *bytes )
+static void CALLBACK dump_callback( unsigned char *user, const struct pcap_pkthdr *hdr, const unsigned char *bytes )
 {
     trace( "user %p hdr %p bytes %p\n", user, hdr, bytes );
     ppcap_dump( user, hdr, bytes );
@@ -314,7 +310,7 @@
     dumper = ppcap_dump_open( pcap, filename );
     ok( dumper != NULL, "got NULL\n" );
 
-    ret = ppcap_dispatch( pcap, 2, dump_callback, (unsigned char *)dumper );
+    ret = ppcap_dispatch( pcap, 1, dump_callback, NULL );
     ok( ret >= 0, "got %d\n", ret );
 
     ppcap_dump_close( dumper );
@@ -365,7 +361,6 @@
     ppcap_loop = (void *)GetProcAddress( module, "pcap_loop" );
     ppcap_set_buffer_size = (void *)GetProcAddress( module, "pcap_set_buffer_size" );
     ppcap_set_datalink = (void *)GetProcAddress( module, "pcap_set_datalink" );
-    ppcap_set_immediate_mode = (void *)GetProcAddress( module, "pcap_set_immediate_mode" );
     ppcap_set_promisc = (void *)GetProcAddress( module, "pcap_set_promisc" );
     ppcap_set_timeout = (void *)GetProcAddress( module, "pcap_set_timeout" );
     ppcap_set_tstamp_precision = (void *)GetProcAddress( module, "pcap_set_tstamp_precision" );
