diff --git a/dlls/ieframe/dochost.c b/dlls/ieframe/dochost.c
--- a/dlls/ieframe/dochost.c
+++ b/dlls/ieframe/dochost.c
@@ -16,8 +16,6 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#include <assert.h>
-
 #include "ieframe.h"
 
 #include "exdispid.h"
@@ -367,11 +365,13 @@
 {
     DocHost *This;
 
+    static const WCHAR wszTHIS[] = {'T','H','I','S',0};
+
     if(msg == WM_CREATE) {
         This = *(DocHost**)lParam;
-        SetPropW(hwnd, L"THIS", This);
+        SetPropW(hwnd, wszTHIS, This);
     }else {
-        This = GetPropW(hwnd, L"THIS");
+        This = GetPropW(hwnd, wszTHIS);
     }
 
     switch(msg) {
@@ -441,7 +441,7 @@
             return;
 
         This->travellog.size = 4;
-    }else if(This->travellog.size < This->travellog.position+2) {
+    }else if(This->travellog.size < This->travellog.position+1) {
         travellog_entry_t *new_travellog;
 
         new_travellog = realloc(This->travellog.log, This->travellog.size * 2 * sizeof(*This->travellog.log));
@@ -469,7 +469,6 @@
 
     if(This->travellog.loading_pos == -1) {
         This->travellog.position++;
-        assert(This->travellog.position < This->travellog.size);
         This->travellog.log[This->travellog.position].stream = NULL;
         This->travellog.log[This->travellog.position].url = NULL;
     }else {
@@ -485,6 +484,9 @@
 void create_doc_view_hwnd(DocHost *This)
 {
     RECT rect;
+
+    static const WCHAR wszShell_DocObject_View[] =
+        {'S','h','e','l','l',' ','D','o','c','O','b','j','e','c','t',' ','V','i','e','w',0};
 
     if(!doc_view_atom) {
         static WNDCLASSEXW wndclass = {
@@ -493,7 +495,7 @@
             doc_view_proc,
             0, 0 /* native uses 4*/, NULL, NULL, NULL,
             (HBRUSH)(COLOR_WINDOW + 1), NULL,
-            L"Shell DocObject View",
+            wszShell_DocObject_View,
             NULL
         };
 
@@ -503,8 +505,8 @@
     }
 
     This->container_vtbl->get_docobj_rect(This, &rect);
-    This->hwnd = CreateWindowExW(0, L"Shell DocObject View",
-         L"Shell DocObject View",
+    This->hwnd = CreateWindowExW(0, wszShell_DocObject_View,
+         wszShell_DocObject_View,
          WS_CHILD | WS_VISIBLE | WS_CLIPSIBLINGS | WS_CLIPCHILDREN | WS_TABSTOP,
          rect.left, rect.top, rect.right, rect.bottom, This->frame_hwnd,
          NULL, ieframe_instance, This);
