diff --git a/dlls/ieframe/intshcut.c b/dlls/ieframe/intshcut.c
--- a/dlls/ieframe/intshcut.c
+++ b/dlls/ieframe/intshcut.c
@@ -82,7 +82,7 @@
 
 static BOOL run_winemenubuilder( const WCHAR *args )
 {
-    static const WCHAR menubuilder[] = L"\\winemenubuilder.exe";
+    static const WCHAR menubuilder[] = {'\\','w','i','n','e','m','e','n','u','b','u','i','l','d','e','r','.','e','x','e',0};
     LONG len;
     LPWSTR buffer;
     STARTUPINFOW si;
@@ -124,7 +124,7 @@
 
 static BOOL StartLinkProcessor( LPCOLESTR szLink )
 {
-    static const WCHAR szFormat[] = L" -w -u \"%s\"";
+    static const WCHAR szFormat[] = { ' ','-','w',' ','-','u',' ','"','%','s','"',0 };
     LONG len;
     LPWSTR buffer;
     BOOL ret;
@@ -260,6 +260,7 @@
     InternetShortcut *This = impl_from_IUniformResourceLocatorW(url);
     WCHAR app[64];
     HKEY hkey;
+    static const WCHAR wszURLProtocol[] = {'U','R','L',' ','P','r','o','t','o','c','o','l',0};
     SHELLEXECUTEINFOW sei;
     DWORD res, type;
     HRESULT hres;
@@ -283,7 +284,7 @@
     if(res != ERROR_SUCCESS)
         return E_FAIL;
 
-    res = RegQueryValueExW(hkey, L"URL Protocol", NULL, &type, NULL, NULL);
+    res = RegQueryValueExW(hkey, wszURLProtocol, NULL, &type, NULL, NULL);
     RegCloseKey(hkey);
     if(res != ERROR_SUCCESS || type != REG_SZ)
         return E_FAIL;
@@ -453,6 +454,10 @@
 
 static HRESULT WINAPI PersistFile_Load(IPersistFile *pFile, LPCOLESTR pszFileName, DWORD dwMode)
 {
+    static const WCHAR str_header[] = {'I','n','t','e','r','n','e','t','S','h','o','r','t','c','u','t',0};
+    static const WCHAR str_URL[] = {'U','R','L',0};
+    static const WCHAR str_iconfile[] = {'i','c','o','n','f','i','l','e',0};
+    static const WCHAR str_iconindex[] = {'i','c','o','n','i','n','d','e','x',0};
     InternetShortcut *This = impl_from_IPersistFile(pFile);
     WCHAR *filename = NULL;
     WCHAR *url;
@@ -470,7 +475,7 @@
     if (!filename)
         return E_OUTOFMEMORY;
 
-    if (FAILED(hr = get_profile_string(L"InternetShortcut", L"URL", pszFileName, &url)))
+    if (FAILED(hr = get_profile_string(str_header, str_URL, pszFileName, &url)))
     {
         CoTaskMemFree(filename);
         return hr;
@@ -495,7 +500,7 @@
        If we don't find them, that's not a failure case -- it's possible
        that they just aren't in there. */
 
-    if (get_profile_string(L"InternetShortcut", L"iconfile", pszFileName, &iconfile) == S_OK)
+    if (get_profile_string(str_header, str_iconfile, pszFileName, &iconfile) == S_OK)
     {
         PROPSPEC ps;
         PROPVARIANT pv;
@@ -509,7 +514,7 @@
     }
     CoTaskMemFree(iconfile);
 
-    if (get_profile_string(L"InternetShortcut", L"iconindex", pszFileName, &iconindexstring) == S_OK)
+    if (get_profile_string(str_header, str_iconindex, pszFileName, &iconindexstring) == S_OK)
     {
         int iconindex;
         PROPSPEC ps;
