diff --git a/dlls/msvcirt/tests/msvcirt.c b/dlls/msvcirt/tests/msvcirt.c
--- a/dlls/msvcirt/tests/msvcirt.c
+++ b/dlls/msvcirt/tests/msvcirt.c
@@ -493,7 +493,7 @@
 /* Emulate a __thiscall */
 #ifdef __i386__
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 struct thiscall_thunk
 {
     BYTE pop_eax;    /* popl  %eax (ret addr) */
@@ -502,7 +502,7 @@
     BYTE push_eax;   /* pushl %eax */
     WORD jmp_edx;    /* jmp  *%edx */
 };
-#pragma pack(pop)
+#include "poppack.h"
 
 static void * (WINAPI *call_thiscall_func1)( void *func, void *this );
 static void * (WINAPI *call_thiscall_func2)( void *func, void *this, void *a );
@@ -2263,15 +2263,6 @@
     CloseHandle(lock_arg.lock);
     CloseHandle(lock_arg.test);
     CloseHandle(thread);
-}
-
-static void *custom_alloc(LONG size)
-{
-    /* this allocates size + 16 bytes. a test later validates that strstreambuf_overflow
-     * always allocates in `strstreambuf.increase` increments, even when the resultant buffer
-     * won't be large enough to contain `pptr`. to avoid actually writing out-of-bound, we
-     * allocate a few bytes more here. */
-    return p_operator_new(size + 16);
 }
 
 static void test_strstreambuf(void)
@@ -2413,7 +2404,7 @@
     ssb2.base.eback = ssb2.base.base;
     ssb2.base.gptr = ssb2.base.base + 2;
     ssb2.base.egptr = ssb2.base.base + 4;
-    memcpy(ssb2.base.base, "Check", 5);
+    strcpy(ssb2.base.base, "Check");
     ret = (int) call_func1(p_strstreambuf_doallocate, &ssb2);
     ok(ret == 1, "return value %d\n", ret);
     ok(ssb2.base.ebuf == ssb2.base.base + 10, "expected %p, got %p\n", ssb2.base.base + 10, ssb2.base.ebuf);
@@ -2541,12 +2532,8 @@
     ok(ssb2.base.pptr == ssb2.base.base + 1, "wrong put pointer, expected %p got %p\n", ssb2.base.base + 1, ssb2.base.pptr);
     ok(ssb2.base.epptr == ssb2.base.base + 16, "wrong put end, expected %p got %p\n", ssb2.base.base + 16, ssb2.base.epptr);
     ok(*(ssb2.base.pptr - 1) == 'D', "expected 'D' got %d\n", *(ssb2.base.pptr - 1));
-    /* test strstreambuf_overflow always allocate in `increase` increments, even when the new size
-     * won't be large enough to contain `pptr`. `custom_alloc` allocate more space than asked for
-     * to avoid out-of-bound access. */
     ssb2.base.pbase = ssb2.base.base + 3;
     ssb2.base.pptr = ssb2.base.epptr + 5;
-    ssb2.f_alloc = custom_alloc;
     ret = (int) call_func2(p_strstreambuf_overflow, &ssb2, 'E');
     ok(ret == 1, "expected 1 got %d\n", ret);
     ok(ssb2.base.ebuf == ssb2.base.base + 20, "expected %p got %p\n", ssb2.base.base + 20, ssb2.base.ebuf);
@@ -2555,7 +2542,6 @@
     ok(ssb2.base.pptr == ssb2.base.base + 22, "wrong put pointer, expected %p got %p\n", ssb2.base.base + 22, ssb2.base.pptr);
     ok(ssb2.base.epptr == ssb2.base.base + 20, "wrong put end, expected %p got %p\n", ssb2.base.base + 20, ssb2.base.epptr);
     ok(*(ssb2.base.pptr - 1) == 'E', "expected 'E' got %d\n", *(ssb2.base.pptr - 1));
-    ssb2.f_alloc = NULL;
     ssb2.base.egptr = ssb2.base.base + 2;
     ret = (int) call_func2(p_strstreambuf_overflow, &ssb2, 'F');
     ok(ret == 1, "expected 1 got %d\n", ret);
@@ -7394,7 +7380,7 @@
     call_func5(p_ifstream_buffer_ctor, &ifs, -1, NULL, 0, TRUE);
     ok(ifs.base_ios.sb->base == NULL, "wrong base value, expected NULL got %p\n", ifs.base_ios.sb->base);
     ok(ifs.base_ios.sb->ebuf == NULL, "wrong ebuf value, expected NULL got %p\n", ifs.base_ios.sb->ebuf);
-    ok(ifs.base_ios.sb->unbuffered == 1, "wrong unbuffered value, expected 1 got %d\n", ifs.base_ios.sb->unbuffered);
+    ok(ifs.base_ios.sb->unbuffered == 1, "wrong unbuffered value, expected 1 got %d\n", pfb->base.unbuffered);
     ok(ifs.base_ios.sb->allocated == 0, "wrong allocated value, expected 0 got %d\n", ifs.base_ios.sb->allocated);
 
     psb = call_func3(p_ifstream_setbuf, &ifs, buffer, ARRAY_SIZE(buffer));
