diff --git a/dlls/dmime/midi.c b/dlls/dmime/midi.c
--- a/dlls/dmime/midi.c
+++ b/dlls/dmime/midi.c
@@ -146,32 +146,6 @@
 
         switch (event->meta_type)
         {
-        case MIDI_META_TEXT_EVENT:
-        case MIDI_META_COPYRIGHT_NOTICE:
-        case MIDI_META_TRACK_NAME:
-        case MIDI_META_INSTRUMENT_NAME:
-        case MIDI_META_LYRIC:
-        case MIDI_META_MARKER:
-        case MIDI_META_CUE_POINT:
-        {
-            char *str = malloc(length);
-            if (FAILED(hr = stream_read_at_most(stream, str, length, bytes_left)))
-            {
-                free(str);
-                return hr;
-            }
-            TRACE("MIDI meta event type %#02x, text: %s\n", event->meta_type, debugstr_an(str, length));
-            free(str);
-            /* Skip over this event */
-            return read_midi_event(stream, event, last_status, bytes_left);
-        }
-        case MIDI_META_END_OF_TRACK:
-            if (length)
-            {
-                ERR("Invalid MIDI meta event length %lu for end of track event.\n", length);
-                return E_FAIL;
-            }
-            break;
         case MIDI_META_SET_TEMPO:
             if (length != 3)
             {
@@ -181,18 +155,6 @@
             if (FAILED(hr = stream_read_at_most(stream, data, 3, bytes_left))) return hr;
             event->tempo = (data[0] << 16) | (data[1] << 8) | data[2];
             break;
-        case MIDI_META_KEY_SIGNATURE:
-            if (length != 2)
-            {
-                ERR("Invalid MIDI meta event length %lu for key signature event.\n", length);
-                return E_FAIL;
-            }
-            if (FAILED(hr = stream_read_at_most(stream, data, 2, bytes_left))) return hr;
-            TRACE("MIDI key signature meta event: %d %s, %s key.\n", data[0],
-                    data[0] < 0 ? "flats" : (data[0] > 0 ? "sharps" : "no flats nor sharps"),
-                    data[1] == 0 ? "major" : "minor");
-            /* Skip over this event */
-            return read_midi_event(stream, event, last_status, bytes_left);
         default:
             if (*bytes_left < length) return S_FALSE;
             offset.QuadPart = length;
@@ -406,20 +368,8 @@
         while ((hr = read_midi_event(parser->stream, &event, &last_status, &length)) == S_OK)
         {
             parser->time += event.delta_time;
-            if (event.status == MIDI_META)
-            {
-                if (event.meta_type == MIDI_META_END_OF_TRACK)
-                    break;
-                switch (event.meta_type)
-                {
-                case MIDI_META_SET_TEMPO:
-                    hr = midi_parser_handle_set_tempo(parser, &event);
-                    break;
-                default:
-                    FIXME("Unhandled MIDI meta event type %#02x at time +%lu\n", event.meta_type, parser->time);
-                    break;
-                }
-            }
+            if (event.status == 0xff && event.meta_type == MIDI_META_SET_TEMPO)
+                hr = midi_parser_handle_set_tempo(parser, &event);
             else
             {
                 switch (event.status & 0xf0)
