diff --git a/dlls/gdiplus/imageattributes.c b/dlls/gdiplus/imageattributes.c
--- a/dlls/gdiplus/imageattributes.c
+++ b/dlls/gdiplus/imageattributes.c
@@ -27,11 +27,6 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(gdiplus);
 
-static inline int is_valid_ColorAdjustType(ColorAdjustType type)
-{
-    return (type >= ColorAdjustTypeDefault && type < ColorAdjustTypeCount);
-}
-
 GpStatus WINGDIPAPI GdipCloneImageAttributes(GDIPCONST GpImageAttributes *imageattr,
     GpImageAttributes **cloneImageattr)
 {
@@ -122,7 +117,7 @@
     TRACE("(%p,%p,%u)\n", imageattr, palette, type);
 
     if (!imageattr || !palette || !palette->Count ||
-        !is_valid_ColorAdjustType(type) || type == ColorAdjustTypeDefault)
+        type >= ColorAdjustTypeCount || type == ColorAdjustTypeDefault)
         return InvalidParameter;
 
     apply_image_attributes(imageattr, (LPBYTE)palette->Entries, palette->Count, 1, 0,
@@ -136,7 +131,7 @@
 {
     TRACE("(%p,%u,%i,%08lx,%08lx)\n", imageattr, type, enableFlag, colorLow, colorHigh);
 
-    if(!imageattr || !is_valid_ColorAdjustType(type))
+    if(!imageattr || type >= ColorAdjustTypeCount)
         return InvalidParameter;
 
     imageattr->colorkeys[type].enabled = enableFlag;
@@ -153,7 +148,7 @@
     TRACE("(%p,%u,%i,%p,%p,%u)\n", imageattr, type, enableFlag, colorMatrix,
         grayMatrix, flags);
 
-    if(!imageattr || !is_valid_ColorAdjustType(type) || flags > ColorMatrixFlagsAltGray)
+    if(!imageattr || type >= ColorAdjustTypeCount || flags > ColorMatrixFlagsAltGray)
         return InvalidParameter;
 
     if (enableFlag)
@@ -211,7 +206,7 @@
 {
     TRACE("(%p,%u,%i,%0.2f)\n", imageAttr, type, enableFlag, gamma);
 
-    if (!imageAttr || (enableFlag && gamma <= 0.0) || !is_valid_ColorAdjustType(type))
+    if (!imageAttr || (enableFlag && gamma <= 0.0) || type >= ColorAdjustTypeCount)
         return InvalidParameter;
 
     imageAttr->gamma_enabled[type] = enableFlag;
@@ -225,7 +220,7 @@
 {
     TRACE("(%p,%u,%i)\n", imageAttr, type, enableFlag);
 
-    if (!is_valid_ColorAdjustType(type))
+    if (type >= ColorAdjustTypeCount)
         return InvalidParameter;
 
     imageAttr->noop[type] = enableFlag ? IMAGEATTR_NOOP_SET : IMAGEATTR_NOOP_CLEAR;
@@ -268,7 +263,7 @@
 
     TRACE("(%p,%u,%i,%u,%p)\n", imageAttr, type, enableFlag, mapSize, map);
 
-    if(!imageAttr || !is_valid_ColorAdjustType(type))
+    if(!imageAttr || type >= ColorAdjustTypeCount)
 	return InvalidParameter;
 
     if (enableFlag)
@@ -330,7 +325,7 @@
 {
     TRACE("(%p,%u)\n", imageAttr, type);
 
-    if(!imageAttr || !is_valid_ColorAdjustType(type))
+    if(!imageAttr || type >= ColorAdjustTypeCount)
         return InvalidParameter;
 
     memset(&imageAttr->colormatrices[type], 0, sizeof(imageAttr->colormatrices[type]));
