diff --git a/dlls/user.exe16/message.c b/dlls/user.exe16/message.c
--- a/dlls/user.exe16/message.c
+++ b/dlls/user.exe16/message.c
@@ -34,8 +34,6 @@
 WINE_DEFAULT_DEBUG_CHANNEL(msg);
 WINE_DECLARE_DEBUG_CHANNEL(message);
 
-#define MAX_ATOM_LEN 255
-
 DWORD USER16_AlertableWait = 0;
 
 struct wow_handlers32 wow_handlers32;
@@ -101,7 +99,7 @@
  * Support for window procedure thunks
  */
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 typedef struct
 {
     WORD        popl_eax;        /* popl  %eax (return address) */
@@ -111,7 +109,7 @@
     BYTE        ljmp;            /* ljmp relay*/
     FARPROC16   relay;           /* __wine_call_wndproc */
 } WINPROC_THUNK;
-#pragma pack(pop)
+#include "poppack.h"
 
 #define WINPROC_HANDLE (~0u >> 16)
 #define MAX_WINPROCS32 4096
@@ -322,7 +320,7 @@
 
 #define MAX_THUNKS 32
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 static struct word_break_thunk
 {
     BYTE                popl_eax;       /* popl  %eax (return address) */
@@ -332,7 +330,7 @@
     BYTE                jmp;            /* ljmp call_word_break_proc16 */
     DWORD               callback;
 } *word_break_thunks;
-#pragma pack(pop)
+#include <poppack.h>
 
 /**********************************************************************
  *           call_word_break_proc16
@@ -883,9 +881,9 @@
             UINT_PTR lo = LOWORD(lParam);
             UINT_PTR hi = HIWORD(lParam);
             int flag = 0;
-            char buf[MAX_ATOM_LEN + 1];
-
-            if (GlobalGetAtomNameA(hi, buf, sizeof(buf)) > 0) flag |= 1;
+            char buf[2];
+
+            if (GlobalGetAtomNameA(hi, buf, 2) > 0) flag |= 1;
             if (GlobalSize16(hi) != 0) flag |= 2;
             switch (flag)
             {
@@ -1264,14 +1262,9 @@
         {
             UINT_PTR lo, hi;
             int flag = 0;
-            char buf[MAX_ATOM_LEN + 1];
-
-            if (!UnpackDDElParam( msg, lParam, &lo, &hi ))
-            {
-                /* Probably this is a response to WM_DDE_INITIATE */
-                lo = LOWORD( lParam );
-                hi = HIWORD( lParam );
-            }
+            char buf[2];
+
+            UnpackDDElParam( msg, lParam, &lo, &hi );
 
             if (GlobalGetAtomNameA((ATOM)hi, buf, sizeof(buf)) > 0) flag |= 1;
             if (GlobalSize((HANDLE)hi) != 0) flag |= 2;
