diff --git a/dlls/msvcr70/tests/msvcr70.c b/dlls/msvcr70/tests/msvcr70.c
--- a/dlls/msvcr70/tests/msvcr70.c
+++ b/dlls/msvcr70/tests/msvcr70.c
@@ -25,38 +25,63 @@
 #include <errno.h>
 #include "wine/test.h"
 
+static int (__cdecl *p_strcmp)(const char *, const char *);
+static int (__cdecl *p_strncmp)(const char *, const char *, size_t);
+
+#define SETNOFAIL(x,y) x = (void*)GetProcAddress(hcrt,y)
+#define SET(x,y) do { SETNOFAIL(x,y); ok(x != NULL, "Export '%s' not found\n", y); } while(0)
+static BOOL init(void)
+{
+    HMODULE hcrt;
+
+    SetLastError(0xdeadbeef);
+    hcrt = LoadLibraryA("msvcr70.dll");
+    if (!hcrt) {
+        win_skip("msvcr70.dll not installed (got %ld)\n", GetLastError());
+        return FALSE;
+    }
+
+    SET(p_strcmp, "strcmp");
+    SET(p_strncmp, "strncmp");
+
+    return TRUE;
+}
+
 static void test_strcmp(void)
 {
-    int ret = strcmp( "abc", "abcd" );
+    int ret = p_strcmp( "abc", "abcd" );
     ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strcmp( "", "abc" );
+    ret = p_strcmp( "", "abc" );
     ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strcmp( "abc", "ab\xa0" );
+    ret = p_strcmp( "abc", "ab\xa0" );
     ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strcmp( "ab\xb0", "ab\xa0" );
+    ret = p_strcmp( "ab\xb0", "ab\xa0" );
     ok( ret == 1, "wrong ret %d\n", ret );
-    ret = strcmp( "ab\xc2", "ab\xc2" );
+    ret = p_strcmp( "ab\xc2", "ab\xc2" );
     ok( ret == 0, "wrong ret %d\n", ret );
 
-    ret = strncmp( "abc", "abcd", 3 );
+    ret = p_strncmp( "abc", "abcd", 3 );
     ok( ret == 0, "wrong ret %d\n", ret );
-    ret = strncmp( "", "abc", 3 );
+    ret = p_strncmp( "", "abc", 3 );
     ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strncmp( "abc", "ab\xa0", 4 );
+    ret = p_strncmp( "abc", "ab\xa0", 4 );
     ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strncmp( "ab\xb0", "ab\xa0", 3 );
+    ret = p_strncmp( "ab\xb0", "ab\xa0", 3 );
     ok( ret == 1, "wrong ret %d\n", ret );
-    ret = strncmp( "ab\xb0", "ab\xa0", 2 );
+    ret = p_strncmp( "ab\xb0", "ab\xa0", 2 );
     ok( ret == 0, "wrong ret %d\n", ret );
-    ret = strncmp( "ab\xc2", "ab\xc2", 3 );
+    ret = p_strncmp( "ab\xc2", "ab\xc2", 3 );
     ok( ret == 0, "wrong ret %d\n", ret );
-    ret = strncmp( "abc", "abd", 0 );
+    ret = p_strncmp( "abc", "abd", 0 );
     ok( ret == 0, "wrong ret %d\n", ret );
-    ret = strncmp( "abc", "abc", 12 );
+    ret = p_strncmp( "abc", "abc", 12 );
     ok( ret == 0, "wrong ret %d\n", ret );
 }
 
 START_TEST(msvcr70)
 {
+    if(!init())
+        return;
+
     test_strcmp();
 }
