diff --git a/dlls/ntoskrnl.exe/sync.c b/dlls/ntoskrnl.exe/sync.c
--- a/dlls/ntoskrnl.exe/sync.c
+++ b/dlls/ntoskrnl.exe/sync.c
@@ -23,6 +23,7 @@
 #include "ntoskrnl_private.h"
 #include "ddk/ntddk.h"
 
+#include "wine/heap.h"
 #include "wine/server.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(ntoskrnl);
@@ -97,7 +98,7 @@
     }
     LeaveCriticalSection( &sync_cs );
 
-    ret = NtWaitForMultipleObjects( count, handles, wait_type, alertable, timeout );
+    ret = NtWaitForMultipleObjects( count, handles, (wait_type == WaitAny), alertable, timeout );
 
     EnterCriticalSection( &sync_cs );
     for (i = 0; i < count; i++)
@@ -909,9 +910,9 @@
 NTSTATUS WINAPI ExDeleteResourceLite( ERESOURCE *resource )
 {
     TRACE("resource %p.\n", resource);
-    HeapFree(GetProcessHeap(), 0, resource->OwnerTable);
-    HeapFree(GetProcessHeap(), 0, resource->ExclusiveWaiters);
-    HeapFree(GetProcessHeap(), 0, resource->SharedWaiters);
+    heap_free(resource->OwnerTable);
+    heap_free(resource->ExclusiveWaiters);
+    heap_free(resource->SharedWaiters);
     return STATUS_SUCCESS;
 }
 
@@ -927,11 +928,7 @@
     }
 
     count = ++resource->OwnerEntry.TableSize;
-    if (!resource->OwnerTable)
-        resource->OwnerTable = HeapAlloc(GetProcessHeap(), 0, count * sizeof(*resource->OwnerTable));
-    else
-        resource->OwnerTable = HeapReAlloc(GetProcessHeap(), 0, resource->OwnerTable,
-                                            count * sizeof(*resource->OwnerTable));
+    resource->OwnerTable = heap_realloc(resource->OwnerTable, count * sizeof(*resource->OwnerTable));
     resource->OwnerTable[count - 1].OwnerThread = thread;
     resource->OwnerTable[count - 1].OwnerCount = 0;
 
@@ -974,7 +971,7 @@
 
     if (!resource->ExclusiveWaiters)
     {
-        resource->ExclusiveWaiters = HeapAlloc( GetProcessHeap(), 0, sizeof(*resource->ExclusiveWaiters) );
+        resource->ExclusiveWaiters = heap_alloc( sizeof(*resource->ExclusiveWaiters) );
         KeInitializeEvent( resource->ExclusiveWaiters, SynchronizationEvent, FALSE );
     }
     resource->NumberOfExclusiveWaiters++;
@@ -1037,7 +1034,7 @@
 
     if (!resource->SharedWaiters)
     {
-        resource->SharedWaiters = HeapAlloc( GetProcessHeap(), 0, sizeof(*resource->SharedWaiters) );
+        resource->SharedWaiters = heap_alloc( sizeof(*resource->SharedWaiters) );
         KeInitializeSemaphore( resource->SharedWaiters, 0, INT_MAX );
     }
     resource->NumberOfSharedWaiters++;
@@ -1101,7 +1098,7 @@
 
     if (!resource->SharedWaiters)
     {
-        resource->SharedWaiters = HeapAlloc( GetProcessHeap(), 0, sizeof(*resource->SharedWaiters) );
+        resource->SharedWaiters = heap_alloc( sizeof(*resource->SharedWaiters) );
         KeInitializeSemaphore( resource->SharedWaiters, 0, INT_MAX );
     }
     resource->NumberOfSharedWaiters++;
@@ -1163,7 +1160,7 @@
 
     if (!resource->SharedWaiters)
     {
-        resource->SharedWaiters = HeapAlloc( GetProcessHeap(), 0, sizeof(*resource->SharedWaiters) );
+        resource->SharedWaiters = heap_alloc( sizeof(*resource->SharedWaiters) );
         KeInitializeSemaphore( resource->SharedWaiters, 0, INT_MAX );
     }
     resource->NumberOfSharedWaiters++;
