diff --git a/dlls/crypt32/cert.c b/dlls/crypt32/cert.c
--- a/dlls/crypt32/cert.c
+++ b/dlls/crypt32/cert.c
@@ -212,7 +212,6 @@
         if (existing)
         {
             TRACE("found matching certificate, not adding\n");
-            CertFreeCertificateContext(existing);
             SetLastError(CRYPT_E_EXISTS);
             return FALSE;
         }
@@ -232,9 +231,7 @@
         {
             Context_CopyProperties(existing, cert);
             if (ret_context)
-                *ret_context = existing;
-            else
-                CertFreeCertificateContext(existing);
+                *ret_context = CertDuplicateCertificateContext(existing);
             return TRUE;
         }
         break;
@@ -242,7 +239,6 @@
         if (existing && CompareFileTime(&existing->pCertInfo->NotBefore, &cert->pCertInfo->NotBefore) >= 0)
         {
             TRACE("existing certificate is newer, not adding\n");
-            CertFreeCertificateContext(existing);
             SetLastError(CRYPT_E_EXISTS);
             return FALSE;
         }
@@ -253,7 +249,6 @@
             if (CompareFileTime(&existing->pCertInfo->NotBefore, &cert->pCertInfo->NotBefore) >= 0)
             {
                 TRACE("existing certificate is newer, not adding\n");
-                CertFreeCertificateContext(existing);
                 SetLastError(CRYPT_E_EXISTS);
                 return FALSE;
             }
@@ -272,10 +267,7 @@
     ret = store->vtbl->certs.addContext(store, context_from_ptr(cert), existing ? context_from_ptr(existing) : NULL,
      (ret_context || inherit_props) ? &new_context : NULL, use_link);
     if(!ret)
-    {
-        CertFreeCertificateContext(existing);
         return FALSE;
-    }
 
     if(inherit_props)
         Context_CopyProperties(context_ptr(new_context), existing);
@@ -284,9 +276,6 @@
         *ret_context = context_ptr(new_context);
     else if(new_context)
         Context_Release(new_context);
-
-    if (existing)
-        CertFreeCertificateContext(existing);
 
     TRACE("returning %d\n", ret);
     return ret;
@@ -879,7 +868,7 @@
 static BOOL CRYPT_AcquirePrivateKeyFromProvInfo(PCCERT_CONTEXT pCert, DWORD dwFlags,
  PCRYPT_KEY_PROV_INFO info, HCRYPTPROV *phCryptProv, DWORD *pdwKeySpec)
 {
-    DWORD size = 0, flags = (dwFlags & CRYPT_ACQUIRE_SILENT_FLAG) ? CRYPT_SILENT : 0;
+    DWORD size = 0;
     BOOL allocated = FALSE, ret = TRUE;
 
     if (!info)
@@ -906,12 +895,8 @@
     }
     if (ret)
     {
-        ret = CryptAcquireContextW(phCryptProv, info->pwszContainerName, info->pwszProvName, info->dwProvType, flags);
-        if (!ret)
-        {
-            flags |= CRYPT_MACHINE_KEYSET;
-            ret = CryptAcquireContextW(phCryptProv, info->pwszContainerName, info->pwszProvName, info->dwProvType, flags);
-        }
+        ret = CryptAcquireContextW(phCryptProv, info->pwszContainerName,
+         info->pwszProvName, info->dwProvType, (dwFlags & CRYPT_ACQUIRE_SILENT_FLAG) ? CRYPT_SILENT : 0);
         if (ret)
         {
             DWORD i;
