diff --git a/dlls/wmphoto/jxrlib.c b/dlls/wmphoto/jxrlib.c
--- a/dlls/wmphoto/jxrlib.c
+++ b/dlls/wmphoto/jxrlib.c
@@ -249,11 +249,6 @@
     return S_OK;
 }
 
-static HRESULT CDECL wmp_decoder_get_decoder_palette(struct decoder *iface, UINT frame, WICColor *colors, UINT *num_colors)
-{
-    return WINCODEC_ERR_PALETTEUNAVAILABLE;
-}
-
 static HRESULT CDECL wmp_decoder_copy_pixels(struct decoder *iface, UINT frame, const WICRect *prc, UINT stride, UINT buffersize, BYTE *buffer)
 {
     struct wmp_decoder *This = impl_from_decoder(iface);
@@ -275,12 +270,12 @@
         pkrect.Width = This->frame.width;
         pkrect.Height = This->frame.height;
         This->frame_stride = (This->frame.width * This->frame.bpp + 7) / 8;
-        if (!(frame_data = calloc(This->frame.height, This->frame_stride)))
+        if (!(frame_data = RtlAllocateHeap(GetProcessHeap(), HEAP_ZERO_MEMORY, This->frame.height * This->frame_stride)))
             return E_FAIL;
         if (This->decoder->Copy(This->decoder, &pkrect, frame_data, stride))
         {
             ERR("Failed to copy frame data!\n");
-            free(frame_data);
+            RtlFreeHeap(GetProcessHeap(), 0, frame_data);
             return E_FAIL;
         }
 
@@ -319,14 +314,14 @@
     }
     *datasize = count;
 
-    bytes = malloc(count);
+    bytes = RtlAllocateHeap(GetProcessHeap(), 0, count);
     if (!bytes)
         return E_OUTOFMEMORY;
 
     if (This->decoder->GetColorContext(This->decoder, bytes, &count))
     {
         ERR("Failed to get frame color context!\n");
-        free(bytes);
+        RtlFreeHeap(GetProcessHeap(), 0, bytes);
         return E_FAIL;
     }
 
@@ -341,14 +336,13 @@
     TRACE("iface %p\n", iface);
 
     This->decoder->Release(&This->decoder);
-    free(This->frame_data);
-    free(This);
+    RtlFreeHeap(GetProcessHeap(), 0, This->frame_data);
+    RtlFreeHeap(GetProcessHeap(), 0, This);
 }
 
 static const struct decoder_funcs wmp_decoder_vtable = {
     wmp_decoder_initialize,
     wmp_decoder_get_frame_info,
-    wmp_decoder_get_decoder_palette,
     wmp_decoder_copy_pixels,
     wmp_decoder_get_metadata_blocks,
     wmp_decoder_get_color_context,
@@ -361,7 +355,7 @@
     PKImageDecode *decoder;
 
     if (PKImageDecode_Create_WMP(&decoder)) return E_FAIL;
-    This = calloc(1, sizeof(*This));
+    This = RtlAllocateHeap(GetProcessHeap(), 0, sizeof(*This));
     if (!This)
     {
         decoder->Release(&decoder);
@@ -377,6 +371,10 @@
     This->WMPStream_iface.GetPos = &wmp_stream_GetPos;
 
     This->decoder = decoder;
+    This->stream = NULL;
+    memset(&This->frame, 0, sizeof(This->frame));
+    This->frame_stride = 0;
+    This->frame_data = NULL;
 
     *result = &This->decoder_iface;
 
