diff --git a/dlls/dinput/device.c b/dlls/dinput/device.c
--- a/dlls/dinput/device.c
+++ b/dlls/dinput/device.c
@@ -415,17 +415,16 @@
     /* Try to read each action in the DIACTIONFORMAT from registry */
     for (i = 0; i < lpdiaf->dwNumActions; i++)
     {
-        DIACTIONW *action = lpdiaf->rgoAction + i;
         DWORD id, size = sizeof(DWORD);
         WCHAR label[9];
 
         swprintf( label, 9, L"%x", lpdiaf->rgoAction[i].dwSemantic );
 
-        if (!action->dwHow && !RegQueryValueExW( hkey, label, 0, NULL, (BYTE *)&id, &size ))
-        {
-            action->dwObjID = id;
-            action->guidInstance = didev.guidInstance;
-            action->dwHow = DIAH_DEFAULT;
+        if (!RegQueryValueExW(hkey, label, 0, NULL, (LPBYTE) &id, &size))
+        {
+            lpdiaf->rgoAction[i].dwObjID = id;
+            lpdiaf->rgoAction[i].guidInstance = didev.guidInstance;
+            lpdiaf->rgoAction[i].dwHow = DIAH_DEFAULT;
             mapped += 1;
         }
     }
@@ -1803,7 +1802,7 @@
 BOOL device_object_matches_semantic( const DIDEVICEINSTANCEW *instance, const DIOBJECTDATAFORMAT *object,
                                      DWORD semantic, BOOL exact )
 {
-    DWORD value = semantic & 0xff, axis = (semantic >> 15) & 15, type;
+    DWORD value = semantic & 0xff, axis = (semantic >> 15) & 3, type;
 
     switch (semantic & 0x700)
     {
@@ -1815,20 +1814,17 @@
     }
 
     if (!(DIDFT_GETTYPE( object->dwType ) & type)) return FALSE;
-    switch (semantic & 0xff000000)
-    {
-    case 0x81000000: return (instance->dwDevType & 0xf) == DIDEVTYPE_KEYBOARD && object->dwOfs == value;
-    case 0x82000000: return (instance->dwDevType & 0xf) == DIDEVTYPE_MOUSE && object->dwOfs == value;
-    case 0x83000000: return FALSE;
-
-    default:
-        if ((instance->dwDevType & 0xf) == DIDEVTYPE_KEYBOARD) return FALSE;
-        if ((instance->dwDevType & 0xf) == DIDEVTYPE_MOUSE) return FALSE;
-        /* fallthrough */
-    case 0xff000000:
-        if (axis && (axis - 1) != DIDFT_GETINSTANCE( object->dwType )) return FALSE;
-        return !exact || !value || value == DIDFT_GETINSTANCE( object->dwType ) + 1;
-    }
+    if ((semantic & 0xf0000000) == 0x80000000)
+    {
+        switch (semantic & 0x0f000000)
+        {
+        case 0x01000000: return (instance->dwDevType & 0xf) == DIDEVTYPE_KEYBOARD && object->dwOfs == value;
+        case 0x02000000: return (instance->dwDevType & 0xf) == DIDEVTYPE_MOUSE && object->dwOfs == value;
+        default: return FALSE;
+        }
+    }
+    if (axis && (axis - 1) != DIDFT_GETINSTANCE( object->dwType )) return FALSE;
+    return !exact || !value || value == DIDFT_GETINSTANCE( object->dwType ) + 1;
 }
 
 static HRESULT WINAPI dinput_device_BuildActionMap( IDirectInputDevice8W *iface, DIACTIONFORMATW *format,
@@ -1867,14 +1863,11 @@
         if (!action->dwSemantic) return DIERR_INVALIDPARAM;
         if (flags == DIDBAM_PRESERVE && !IsEqualCLSID( &action->guidInstance, &GUID_NULL ) &&
             !IsEqualCLSID( &action->guidInstance, &impl->guid )) continue;
-        if (action->dwFlags & DIA_APPMAPPED)
-        {
-            action->dwHow = DIAH_APPREQUESTED;
-            continue;
-        }
+        if (action->dwFlags & DIA_APPMAPPED) action->dwHow = DIAH_APPREQUESTED;
+        else action->dwHow = 0;
+        if (action->dwHow == DIAH_APPREQUESTED || action->dwHow == DIAH_USERCONFIG) continue;
         if ((action->dwSemantic & 0xf0000000) == 0x80000000) action->dwFlags &= ~DIA_APPNOMAP;
         if (!(action->dwFlags & DIA_APPNOMAP)) action->guidInstance = GUID_NULL;
-        action->dwHow = 0;
     }
 
     /* Unless asked the contrary by these flags, try to load a previous mapping */
