diff --git a/dlls/comctl32/imagelist.c b/dlls/comctl32/imagelist.c
--- a/dlls/comctl32/imagelist.c
+++ b/dlls/comctl32/imagelist.c
@@ -71,7 +71,6 @@
     INT         nOvlIdx[MAX_OVERLAYIMAGE]; /* 38: overlay images index */
 
     /* not yet found out */
-    DWORD   magic;
     HBRUSH  hbrBlend25;
     HBRUSH  hbrBlend50;
     INT     cInitial;
@@ -85,7 +84,7 @@
 #define IMAGELIST_MAGIC 0x53414D58
 
 /* Header used by ImageList_Read() and ImageList_Write() */
-#pragma pack(push,2)
+#include "pshpack2.h"
 typedef struct _ILHEAD
 {
     USHORT	usMagic;
@@ -99,7 +98,7 @@
     WORD	flags;
     SHORT	ovls[4];
 } ILHEAD;
-#pragma pack(pop)
+#include "poppack.h"
 
 /* internal image list data used for Drag & Drop operations */
 typedef struct
@@ -2971,7 +2970,6 @@
     return result;
 }
 
-#if __WINE_COMCTL32_VERSION == 6
 /*************************************************************************
  * ImageList_WriteEx [COMCTL32.@]
  */
@@ -2980,7 +2978,6 @@
     FIXME("%p %#lx %p: semi-stub\n", himl, flags, pstm);
     return ImageList_Write(himl, pstm) ? S_OK : E_FAIL;
 }
-#endif /* __WINE_COMCTL32_VERSION == 6 */
 
 /*************************************************************************
  * ImageList_Write [COMCTL32.@]
@@ -3126,7 +3123,6 @@
     return SetDIBColorTable(himl->hdcImage, uStartIndex, cEntries, prgb);
 }
 
-#if __WINE_COMCTL32_VERSION == 6
 /*************************************************************************
  * ImageList_CoCreateInstance [COMCTL32.@]
  *
@@ -3152,7 +3148,6 @@
 
     return ImageListImpl_CreateInstance(punkOuter, riid, ppv);
 }
-#endif /* __WINE_COMCTL32_VERSION == 6 */
 
 
 /*************************************************************************
@@ -3215,7 +3210,6 @@
         if (This->hbrBlend50) DeleteObject (This->hbrBlend50);
 
         This->IImageList2_iface.lpVtbl = NULL;
-        This->magic = 0;
         Free(This->item_flags);
         Free(This);
     }
@@ -3799,7 +3793,7 @@
     BOOL valid;
     __TRY
     {
-        valid = himl && himl->magic == IMAGELIST_MAGIC;
+        valid = himl && himl->IImageList2_iface.lpVtbl == &ImageListImpl_Vtbl;
     }
     __EXCEPT_PAGE_FAULT
     {
@@ -3846,7 +3840,6 @@
     if (!This) return E_OUTOFMEMORY;
 
     This->IImageList2_iface.lpVtbl = &ImageListImpl_Vtbl;
-    This->magic = IMAGELIST_MAGIC;
     This->ref = 1;
 
     ret = IImageList2_QueryInterface(&This->IImageList2_iface, iid, ppv);
