diff --git a/dlls/winhttp/session.c b/dlls/winhttp/session.c
--- a/dlls/winhttp/session.c
+++ b/dlls/winhttp/session.c
@@ -1546,7 +1546,7 @@
         if (err == ERROR_SUCCESS && param.nBytesData)
         {
             int len = MultiByteToWideChar( CP_ACP, 0, (const char *)param.Data, param.nBytesData, NULL, 0 );
-            if ((ret = GlobalAlloc( 0, (len + 1) * sizeof(WCHAR) )))
+            if ((ret = malloc( (len + 1) * sizeof(WCHAR) )))
             {
                 MultiByteToWideChar( CP_ACP, 0,  (const char *)param.Data, param.nBytesData, ret, len );
                 ret[len] = 0;
@@ -1666,13 +1666,6 @@
  */
 BOOL WINAPI WinHttpDetectAutoProxyConfigUrl( DWORD flags, WCHAR **url )
 {
-    static SRWLOCK apc_cache_lock = SRWLOCK_INIT;
-    static BOOL dhcp_cache_valid, dns_cache_valid;
-    static WCHAR *dhcp_apc_url, *dns_apc_url;
-    static HANDLE notifyaddr_handle = INVALID_HANDLE_VALUE;
-    static OVERLAPPED notifyaddr_ol;
-    DWORD bytes;
-
     TRACE( "%#lx, %p\n", flags, url );
 
     if (!flags || !url)
@@ -1681,53 +1674,14 @@
         return FALSE;
     }
     *url = NULL;
-
-    AcquireSRWLockExclusive( &apc_cache_lock );
-    /* FIXME: use NotifyIpInterfaceChange once that's implemented */
-    if (notifyaddr_handle == INVALID_HANDLE_VALUE)
-        NotifyAddrChange( &notifyaddr_handle, &notifyaddr_ol );
-
-    if ((notifyaddr_handle != INVALID_HANDLE_VALUE) &&
-        GetOverlappedResult( notifyaddr_handle, &notifyaddr_ol, &bytes, FALSE ))
-    {
-        TRACE( "network config changed, invalidating cache\n" );
-        dhcp_cache_valid = dns_cache_valid = FALSE;
-        notifyaddr_handle = INVALID_HANDLE_VALUE;
-        NotifyAddrChange( &notifyaddr_handle, &notifyaddr_ol );
-    }
-
     if (flags & WINHTTP_AUTO_DETECT_TYPE_DHCP)
     {
-        if (!dhcp_cache_valid)
-        {
-            GlobalFree( dhcp_apc_url );
-            dhcp_apc_url = detect_autoproxyconfig_url_dhcp();
-            dhcp_cache_valid = TRUE;
-        }
-
-        if (dhcp_apc_url)
-        {
-            *url = GlobalAlloc( 0, (wcslen( dhcp_apc_url ) + 1) * sizeof(WCHAR) );
-            if (*url) wcscpy( *url, dhcp_apc_url );
-        }
-    }
-    if ((flags & WINHTTP_AUTO_DETECT_TYPE_DNS_A) && !*url)
-    {
-        if (!dns_cache_valid)
-        {
-            GlobalFree( dns_apc_url );
-            dns_apc_url = detect_autoproxyconfig_url_dns();
-            dns_cache_valid = TRUE;
-        }
-
-        if (dns_apc_url)
-        {
-            *url = GlobalAlloc( 0, (wcslen( dns_apc_url ) + 1) * sizeof(WCHAR) );
-            if (*url) wcscpy( *url, dns_apc_url );
-        }
-    }
-    ReleaseSRWLockExclusive( &apc_cache_lock );
-
+        *url = detect_autoproxyconfig_url_dhcp();
+    }
+    if (flags & WINHTTP_AUTO_DETECT_TYPE_DNS_A)
+    {
+        if (!*url) *url = detect_autoproxyconfig_url_dns();
+    }
     if (!*url)
     {
         SetLastError( ERROR_WINHTTP_AUTODETECTION_FAILED );
