diff --git a/dlls/ntdll/signal_i386.c b/dlls/ntdll/signal_i386.c
--- a/dlls/ntdll/signal_i386.c
+++ b/dlls/ntdll/signal_i386.c
@@ -72,18 +72,35 @@
                               PEXCEPTION_HANDLER handler, PEXCEPTION_HANDLER nested_handler );
 
 
-#undef SYSCALL_ENTRY_NtQueryInformationProcess
-#define SYSCALL_ENTRY_NtQueryInformationProcess(id,name,args) \
-    __ASM_STDCALL_FUNC( name, args, \
-                        "movl $(" #id "),%eax\n\t" \
-                        "call *%fs:0xc0\n\t" \
-                        "ret $" #args )
+#ifdef __WINE_PE_BUILD
+
+enum syscall_ids
+{
+#define SYSCALL_ENTRY(id,name,args) __id_##name = id,
+ALL_SYSCALLS32
+#undef SYSCALL_ENTRY
+};
+
+/*******************************************************************
+ *         NtQueryInformationProcess
+ */
+void NtQueryInformationProcess_wrapper(void)
+{
+    asm( ".globl " __ASM_STDCALL("NtQueryInformationProcess", 20) "\n"
+         __ASM_STDCALL("NtQueryInformationProcess", 20) ":\n\t"
+         "movl %0,%%eax\n\t"
+         "call *%%fs:0xc0\n\t"
+         "ret $20"  :: "i" (__id_NtQueryInformationProcess) );
+}
+#define NtQueryInformationProcess syscall_NtQueryInformationProcess
+
+#endif /* __WINE_PE_BUILD */
 
 /*******************************************************************
  *         syscalls
  */
 #define SYSCALL_ENTRY(id,name,args) __ASM_SYSCALL_FUNC( id, name, args )
-ALL_SYSCALLS
+ALL_SYSCALLS32
 DEFINE_SYSCALL_HELPER32()
 #undef SYSCALL_ENTRY
 
@@ -215,6 +232,7 @@
  */
 static inline void save_fpu( CONTEXT *context )
 {
+#ifdef __GNUC__
     struct
     {
         DWORD ControlWord;
@@ -235,6 +253,7 @@
     float_status.StatusWord &= float_status.ControlWord | 0xffffff80;
 
     __asm__ __volatile__( "fldenv %0" : : "m" (float_status) );
+#endif
 }
 
 
@@ -245,6 +264,7 @@
  */
 static inline void save_fpux( CONTEXT *context )
 {
+#ifdef __GNUC__
     /* we have to enforce alignment by hand */
     char buffer[sizeof(XSAVE_FORMAT) + 16];
     XSAVE_FORMAT *state = (XSAVE_FORMAT *)(((ULONG_PTR)buffer + 15) & ~15);
@@ -252,6 +272,7 @@
     context->ContextFlags |= CONTEXT_EXTENDED_REGISTERS;
     __asm__ __volatile__( "fxsave %0" : "=m" (*state) );
     memcpy( context->ExtendedRegisters, state, sizeof(*state) );
+#endif
 }
 
 
@@ -518,7 +539,7 @@
                    "leal -12(%esi),%edi\n\t"
                    /* clear the thread stack */
                    "andl $~0xfff,%edi\n\t"   /* round down to page size */
-                   "movl $0xf000,%ecx\n\t"
+                   "movl $0xf0000,%ecx\n\t"
                    "subl %ecx,%edi\n\t"
                    "movl %edi,%esp\n\t"
                    "xorl %eax,%eax\n\t"
