diff --git a/dlls/win32u/spy.c b/dlls/win32u/spy.c
--- a/dlls/win32u/spy.c
+++ b/dlls/win32u/spy.c
@@ -2345,13 +2345,13 @@
             {
                 LPPOINT point = (LPPOINT) sp_e->lParam;
                 if (point) {
-                    TRACE("lParam point x=%d, y=%d\n", point->x, point->y);
+                    TRACE("lParam point x=%d, y=%d\n", (int)point->x, (int)point->y);
                 }
                 break;
             }
         case SBM_SETRANGE:
             if (!enter && (sp_e->msgnum == SBM_SETRANGE)) break;
-            TRACE("min=%d max=%ld\n", (int)sp_e->wParam, sp_e->lParam);
+            TRACE("min=%d max=%d\n", (INT)sp_e->wParam, (INT)sp_e->lParam);
             break;
         case SBM_GETRANGE:
             if ((enter && (sp_e->msgnum == SBM_GETRANGE)) ||
@@ -2371,14 +2371,14 @@
             if (enter && sp_e->lParam)
             {
                 CHARRANGE *cr = (CHARRANGE *) sp_e->lParam;
-                TRACE("CHARRANGE: cpMin=%d cpMax=%d\n", cr->cpMin, cr->cpMax);
+                TRACE("CHARRANGE: cpMin=%d cpMax=%d\n", (int)cr->cpMin, (int)cr->cpMax);
             }
             break;
         case EM_SETCHARFORMAT:
             if (enter && sp_e->lParam)
             {
                 CHARFORMATW *cf = (CHARFORMATW *) sp_e->lParam;
-                TRACE("CHARFORMAT: dwMask=0x%08x dwEffects=", cf->dwMask);
+                TRACE("CHARFORMAT: dwMask=0x%08x dwEffects=", (int)cf->dwMask);
                 if ((cf->dwMask & CFM_BOLD) && (cf->dwEffects & CFE_BOLD))
                     TRACE(" CFE_BOLD");
                 if ((cf->dwMask & CFM_COLOR) && (cf->dwEffects & CFE_AUTOCOLOR))
@@ -2393,11 +2393,11 @@
                     TRACE(" CFE_UNDERLINE");
                 TRACE("\n");
                 if (cf->dwMask & CFM_SIZE)
-                    TRACE("yHeight=%d\n", cf->yHeight);
+                    TRACE("yHeight=%d\n", (int)cf->yHeight);
                 if (cf->dwMask & CFM_OFFSET)
-                    TRACE("yOffset=%d\n", cf->yOffset);
+                    TRACE("yOffset=%d\n", (int)cf->yOffset);
                 if ((cf->dwMask & CFM_COLOR) && !(cf->dwEffects & CFE_AUTOCOLOR))
-                    TRACE("crTextColor=%x\n", cf->crTextColor);
+                    TRACE("crTextColor=%x\n", (int)cf->crTextColor);
                 TRACE("bCharSet=%x bPitchAndFamily=%x\n", cf->bCharSet, cf->bPitchAndFamily);
                 /* FIXME: we should try to be a bit more intelligent about
                  * whether this is in ANSI or Unicode (it could be either) */
@@ -2442,7 +2442,7 @@
             TRACE("%s %s ex=%08x style=%08x %d,%d %dx%d parent=%p menu=%p inst=%p params=%p\n",
                   unicode ? debugstr_w((LPCWSTR)cs->lpszName) : debugstr_a(cs->lpszName),
                   unicode ? debugstr_w((LPCWSTR)cs->lpszClass) : debugstr_a(cs->lpszClass),
-                  cs->dwExStyle, cs->style, cs->x, cs->y, cs->cx, cs->cy,
+                  (int)cs->dwExStyle, (int)cs->style, cs->x, cs->y, cs->cx, cs->cy,
                   cs->hwndParent, cs->hMenu, cs->hInstance, cs->lpCreateParams);
             break;
         }
@@ -2466,7 +2466,7 @@
             {
                 LPSTYLESTRUCT ss = (LPSTYLESTRUCT) sp_e->lParam;
                 TRACE("STYLESTRUCT: StyleOld=0x%08x, StyleNew=0x%08x\n",
-                      ss->styleOld, ss->styleNew);
+                      (int)ss->styleOld, (int)ss->styleNew);
             }
             break;
         case WM_NCCALCSIZE:
@@ -2516,7 +2516,7 @@
                     }
                     if (dumplen > 0) {
                         q = (UINT *)(pnmh + 1);
-                        SPY_DumpMem ("NM extra", q, dumplen);
+                        SPY_DumpMem ("NM extra", q, (INT)dumplen);
                     }
                 }
                 else
@@ -2625,7 +2625,7 @@
             DWORD tid = get_window_thread( hWnd, NULL );
 
             if (tid == GetCurrentThreadId()) strcpy( taskName, "self" );
-            else snprintf( taskName, sizeof(taskName), "tid %04x", GetCurrentThreadId() );
+            else snprintf( taskName, sizeof(taskName), "tid %04x", (int)GetCurrentThreadId() );
 
             TRACE("%*s(%p) %-16s [%04x] %s sent from %s wp=%08lx lp=%08lx\n",
                   indent, "", hWnd, debugstr_w(sp_e.wnd_name), msg,
