diff --git a/dlls/dsound/buffer.c b/dlls/dsound/buffer.c
--- a/dlls/dsound/buffer.c
+++ b/dlls/dsound/buffer.c
@@ -317,6 +317,7 @@
 
 	This->playflags = flags;
 	if (This->state == STATE_STOPPED) {
+		This->leadin = TRUE;
 		This->state = STATE_STARTING;
 	}
 
@@ -417,7 +418,7 @@
 
 	ReleaseSRWLockShared(&This->lock);
 
-	TRACE("playpos = %ld, writepos = %ld, buflen=%ld (%p, time=%lu)\n",
+	TRACE("playpos = %ld, writepos = %ld, buflen=%ld (%p, time=%ld)\n",
 		playpos?*playpos:-1, writepos?*writepos:-1, This->buflen, This, GetTickCount());
 
 	return DS_OK;
@@ -491,18 +492,11 @@
         IDirectSoundBufferImpl *This = impl_from_IDirectSoundBuffer8(iface);
 	HRESULT hres = DS_OK;
 
-        TRACE("(%p,%lu,%lu,%p,%p,%p,%p,0x%08lx) at %lu\n", This, writecursor, writebytes, lplpaudioptr1,
+        TRACE("(%p,%ld,%ld,%p,%p,%p,%p,0x%08lx) at %ld\n", This, writecursor, writebytes, lplpaudioptr1,
                 audiobytes1, lplpaudioptr2, audiobytes2, flags, GetTickCount());
 
-        if (!audiobytes1 || !lplpaudioptr1)
+        if (!audiobytes1)
             return DSERR_INVALIDPARAM;
-
-        *(LPBYTE*)lplpaudioptr1 = NULL;
-        *audiobytes1 = 0;
-        if (lplpaudioptr2)
-            *(LPBYTE*)lplpaudioptr2 = NULL;
-        if (audiobytes2)
-            *audiobytes2 = 0;
 
         /* when this flag is set, writecursor is meaningless and must be calculated */
 	if (flags & DSBLOCK_FROMWRITECURSOR) {
@@ -539,9 +533,13 @@
 			commit_next_chunk(This);
 		}
 		*audiobytes1 = writebytes;
-		TRACE("Locked %p (%lu bytes) and %p (%lu bytes) writecursor=%lu\n",
+		if (lplpaudioptr2)
+			*(LPBYTE*)lplpaudioptr2 = NULL;
+		if (audiobytes2)
+			*audiobytes2 = 0;
+		TRACE("Locked %p(%li bytes) and %p(%li bytes) writecursor=%ld\n",
 		  *(LPBYTE*)lplpaudioptr1, *audiobytes1, lplpaudioptr2 ? *(LPBYTE*)lplpaudioptr2 : NULL, audiobytes2 ? *audiobytes2: 0, writecursor);
-		TRACE("->%lu\n", writebytes);
+		TRACE("->%ld.0\n",writebytes);
 		This->buffer->lockedbytes += writebytes;
 	} else {
 		DWORD remainder = writebytes + writecursor - This->buflen;
@@ -562,9 +560,7 @@
 			WARN("Overwriting mixing position, case 3\n");
 			commit_next_chunk(This);
 		}
-		TRACE("Locked %p (%lu bytes) and %p (%lu bytes) writecursor=%lu\n",
-                      *(LPBYTE*)lplpaudioptr1, *audiobytes1, lplpaudioptr2 ? *(LPBYTE*)lplpaudioptr2 : NULL,
-                      audiobytes2 ? *audiobytes2: 0, writecursor);
+		TRACE("Locked %p(%li bytes) and %p(%li bytes) writecursor=%ld\n", *(LPBYTE*)lplpaudioptr1, *audiobytes1, lplpaudioptr2 ? *(LPBYTE*)lplpaudioptr2 : NULL, audiobytes2 ? *audiobytes2: 0, writecursor);
 	}
 
 	ReleaseSRWLockShared(&This->lock);
@@ -657,7 +653,7 @@
         IDirectSoundBufferImpl *This = impl_from_IDirectSoundBuffer8(iface), *iter;
 	HRESULT hres = DS_OK;
 
-	TRACE("(%p,%p,%lu,%p,%lu)\n", This, p1, x1, p2, x2);
+	TRACE("(%p,%p,%ld,%p,%ld)\n", This,p1,x1,p2,x2);
 
 	if (!p2)
 		x2 = 0;
@@ -1074,8 +1070,8 @@
 	}
 
 	if (dsbd->dwBufferBytes % dsbd->lpwfxFormat->nBlockAlign)
-		dsb->buflen = dsbd->dwBufferBytes +
-			(dsbd->lpwfxFormat->nBlockAlign -
+		dsb->buflen = dsbd->dwBufferBytes + 
+			(dsbd->lpwfxFormat->nBlockAlign - 
 			(dsbd->dwBufferBytes % dsbd->lpwfxFormat->nBlockAlign));
 	else
 		dsb->buflen = dsbd->dwBufferBytes;
