diff --git a/dlls/msvcp140_atomic_wait/tests/msvcp140_atomic_wait.c b/dlls/msvcp140_atomic_wait/tests/msvcp140_atomic_wait.c
--- a/dlls/msvcp140_atomic_wait/tests/msvcp140_atomic_wait.c
+++ b/dlls/msvcp140_atomic_wait/tests/msvcp140_atomic_wait.c
@@ -70,8 +70,6 @@
 static void (__stdcall *p___std_wait_for_threadpool_work_callbacks)(PTP_WORK, BOOL);
 static BOOL (__stdcall *p___std_atomic_wait_direct)(volatile void*, void*, size_t, DWORD);
 static void (__stdcall *p___std_atomic_notify_one_direct)(void*);
-static void (__stdcall *p___std_execution_wait_on_uchar)(void *, unsigned char);
-static void (__stdcall *p___std_execution_wake_by_address_all)(void *);
 static shared_mutex* (__stdcall *p___std_acquire_shared_mutex_for_instance)(void*);
 static void (__stdcall *p___std_release_shared_mutex_for_instance)(void*);
 static struct tzdb_time_zones * (__stdcall *p___std_tzdb_get_time_zones)(void);
@@ -99,8 +97,6 @@
     SET(p___std_wait_for_threadpool_work_callbacks, "__std_wait_for_threadpool_work_callbacks");
     SET(p___std_atomic_wait_direct, "__std_atomic_wait_direct");
     SET(p___std_atomic_notify_one_direct, "__std_atomic_notify_one_direct");
-    SET(p___std_execution_wait_on_uchar, "__std_execution_wait_on_uchar");
-    SET(p___std_execution_wake_by_address_all, "__std_execution_wake_by_address_all");
     SET(p___std_acquire_shared_mutex_for_instance, "__std_acquire_shared_mutex_for_instance");
     SET(p___std_release_shared_mutex_for_instance, "__std_release_shared_mutex_for_instance");
     SET(p___std_tzdb_get_time_zones, "__std_tzdb_get_time_zones");
@@ -295,58 +291,6 @@
     CloseHandle(thread);
 }
 
-static LONG count;
-static unsigned WINAPI execution_wait_thread(void *arg)
-{
-    unsigned char *phase = arg;
-
-    if(!InterlockedDecrement(&count)) {
-        *phase = 1;
-        p___std_execution_wake_by_address_all(phase);
-    }else {
-        while(!*phase)
-            p___std_execution_wait_on_uchar(phase, 0);
-    }
-
-    while(*phase == 1)
-        p___std_execution_wait_on_uchar(phase, 1);
-
-    InterlockedIncrement(&count);
-    return 0;
-}
-
-static void test___std_execution(void)
-{
-    int i;
-    DWORD ret;
-    HANDLE threads[4];
-    unsigned char phase;
-
-    p___std_execution_wake_by_address_all(NULL);
-    if(0) p___std_execution_wait_on_uchar(NULL, 0); /* crash on windows */
-
-    count = ARRAY_SIZE(threads);
-    phase = 0;
-    for(i=0; i<ARRAY_SIZE(threads); i++) {
-        threads[i] = (HANDLE)_beginthreadex(NULL, 0, execution_wait_thread, &phase, 0, NULL);
-        ok(threads[i] != INVALID_HANDLE_VALUE, "_beginthreadex failed\n");
-    }
-
-    Sleep(0);
-    while(!phase)
-        p___std_execution_wait_on_uchar(&phase, 0);
-    ok(!count, "expected 0, got %ld\n", count);
-
-    phase = 2;
-    p___std_execution_wake_by_address_all(&phase);
-
-    ret = WaitForMultipleObjects(ARRAY_SIZE(threads), threads, TRUE, 1000);
-    ok(ret == WAIT_OBJECT_0, "expected WAIT_OBJECT_0, got %ld\n", ret);
-    ok(count == ARRAY_SIZE(threads), "expected %Iu, got %lu\n", ARRAY_SIZE(threads), count);
-    for(i=0; i<ARRAY_SIZE(threads); i++)
-        CloseHandle(threads[i]);
-}
-
 static void test___std_acquire_shared_mutex_for_instance(void)
 {
     shared_mutex *ret1, *ret2;
@@ -429,7 +373,6 @@
     test___std_parallel_algorithms_hw_threads();
     test_threadpool_work();
     test___std_atomic_wait_direct();
-    test___std_execution();
     test___std_acquire_shared_mutex_for_instance();
     test___std_tzdb();
     FreeLibrary(msvcp);
