diff --git a/dlls/user32/message.c b/dlls/user32/message.c
--- a/dlls/user32/message.c
+++ b/dlls/user32/message.c
@@ -29,7 +29,6 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(msg);
 
-#define MAX_ATOM_LEN  255
 
 /* pack a pointer into a 32/64 portable format */
 static inline ULONGLONG pack_ptr( const void *ptr )
@@ -659,6 +658,15 @@
 
 
 /***********************************************************************
+ *		ReplyMessage  (USER32.@)
+ */
+BOOL WINAPI ReplyMessage( LRESULT result )
+{
+    return NtUserReplyMessage( result );
+}
+
+
+/***********************************************************************
  *		InSendMessage  (USER32.@)
  */
 BOOL WINAPI InSendMessage(void)
@@ -989,33 +997,23 @@
  *		RegisterWindowMessageA (USER32.@)
  *		RegisterWindowMessage (USER.118)
  */
-UINT WINAPI RegisterWindowMessageA( LPCSTR name )
-{
-    WCHAR buf[MAX_ATOM_LEN + 1];
-    UNICODE_STRING str = {.Buffer = buf, .MaximumLength = sizeof(buf)};
-    STRING ansi;
-
-    TRACE( "%s\n", debugstr_a(name) );
-
-    RtlInitAnsiString( &ansi, name );
-    RtlAnsiStringToUnicodeString( &str, &ansi, FALSE );
-    return NtUserRegisterWindowMessage( &str );
+UINT WINAPI RegisterWindowMessageA( LPCSTR str )
+{
+    UINT ret = GlobalAddAtomA(str);
+    TRACE("%s, ret=%x\n", str, ret);
+    return ret;
 }
 
 
 /***********************************************************************
  *		RegisterWindowMessageW (USER32.@)
  */
-UINT WINAPI RegisterWindowMessageW( LPCWSTR name )
-{
-    UNICODE_STRING str;
-
-    TRACE( "%s\n", debugstr_w(name) );
-
-    RtlInitUnicodeString( &str, name );
-    return NtUserRegisterWindowMessage( &str );
-}
-
+UINT WINAPI RegisterWindowMessageW( LPCWSTR str )
+{
+    UINT ret = GlobalAddAtomW(str);
+    TRACE("%s ret=%x\n", debugstr_w(str), ret);
+    return ret;
+}
 
 typedef struct BroadcastParm
 {
@@ -1222,6 +1220,15 @@
 }
 
 
+/***********************************************************************
+ *		KillSystemTimer (USER32.@)
+ */
+BOOL WINAPI KillSystemTimer( HWND hwnd, UINT_PTR id )
+{
+    return NtUserKillSystemTimer( hwnd, id );
+}
+
+
 /**********************************************************************
  *		IsGUIThread  (USER32.@)
  */
