diff --git a/dlls/kernelbase/loader.c b/dlls/kernelbase/loader.c
--- a/dlls/kernelbase/loader.c
+++ b/dlls/kernelbase/loader.c
@@ -152,23 +152,33 @@
 static HMODULE load_library( const UNICODE_STRING *libname, DWORD flags )
 {
     const DWORD unsupported_flags = LOAD_IGNORE_CODE_AUTHZ_LEVEL | LOAD_LIBRARY_REQUIRE_SIGNED_TARGET;
+    const ULONG load_library_search_flags = LOAD_WITH_ALTERED_SEARCH_PATH | LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR
+                | LOAD_LIBRARY_SEARCH_APPLICATION_DIR | LOAD_LIBRARY_SEARCH_USER_DIRS
+                | LOAD_LIBRARY_SEARCH_SYSTEM32 | LOAD_LIBRARY_SEARCH_DEFAULT_DIRS;
     NTSTATUS status;
     HMODULE module;
     WCHAR *load_path, *dummy;
+    DWORD load_flags = 0, search_flags;
+
+    TRACE("%s %#lx.\n", libname ? debugstr_w(libname->Buffer) : "<null>", flags);
 
     if (flags & unsupported_flags) FIXME( "unsupported flag(s) used %#08lx\n", flags );
 
-    if (!set_ntstatus( LdrGetDllPath( libname->Buffer, flags, &load_path, &dummy ))) return 0;
+    search_flags = flags & load_library_search_flags;
+    if (flags & DONT_RESOLVE_DLL_REFERENCES)
+        load_flags |= LDR_DONT_RESOLVE_REFS;
 
     if (flags & (LOAD_LIBRARY_AS_DATAFILE | LOAD_LIBRARY_AS_DATAFILE_EXCLUSIVE |
                  LOAD_LIBRARY_AS_IMAGE_RESOURCE))
     {
+        if (!set_ntstatus( LdrGetDllPath( libname->Buffer, flags, &load_path, &dummy ))) return 0;
         if (LdrGetDllHandleEx( 0, load_path, NULL, libname, &module ))
             load_library_as_datafile( load_path, flags, libname->Buffer, &module );
+        RtlReleasePath( load_path );
     }
     else
     {
-        status = LdrLoadDll( load_path, flags, libname, &module );
+        status = LdrLoadDll( (void *)((ULONG_PTR)search_flags | 1), &load_flags, libname, &module );
         if (!set_ntstatus( status ))
         {
             module = 0;
@@ -176,8 +186,6 @@
                 SetLastError( ERROR_DLL_NOT_FOUND );
         }
     }
-
-    RtlReleasePath( load_path );
     return module;
 }
 
@@ -301,7 +309,7 @@
     UNICODE_STRING name;
     NTSTATUS status;
 
-    if (!module && ((win16_tib = NtCurrentTeb()->Tib.SubSystemTib)) && win16_tib->exe_name)
+    if (!module && (0 && (win16_tib = NtCurrentTeb()->Tib.SubSystemTib)) && win16_tib->exe_name)
     {
         len = min( size, win16_tib->exe_name->Length / sizeof(WCHAR) );
         memcpy( filename, win16_tib->exe_name->Buffer, len * sizeof(WCHAR) );
@@ -559,6 +567,14 @@
         SetLastError( ERROR_INVALID_PARAMETER );
         return 0;
     }
+
+    /* HACK: allow webservices.dll to be shipped together with remote debugger tools. */
+    if (flags == LOAD_LIBRARY_SEARCH_SYSTEM32 && !file && !wcscmp( name, L"webservices.dll" ))
+    {
+        FIXME( "HACK: ignoring LOAD_LIBRARY_SEARCH_SYSTEM32 for webservices.dll\n" );
+        flags = 0;
+    }
+
     RtlInitUnicodeString( &str, name );
     if (str.Length && str.Buffer[str.Length/sizeof(WCHAR) - 1] != ' ') return load_library( &str, flags );
 
@@ -1173,7 +1189,7 @@
  */
 HANDLE WINAPI DECLSPEC_HOTPATCH CreateActCtxW( PCACTCTXW ctx )
 {
-    struct _ACTIVATION_CONTEXT *context;
+    HANDLE context;
 
     TRACE( "%p %08lx\n", ctx, ctx ? ctx->dwFlags : 0 );
 
@@ -1225,7 +1241,7 @@
  */
 BOOL WINAPI DECLSPEC_HOTPATCH GetCurrentActCtx( HANDLE *pcontext )
 {
-    return set_ntstatus( RtlGetActiveActivationContext( (struct _ACTIVATION_CONTEXT **)pcontext ));
+    return set_ntstatus( RtlGetActiveActivationContext( pcontext ));
 }
 
 
