diff --git a/dlls/ntdll/tests/string.c b/dlls/ntdll/tests/string.c
--- a/dlls/ntdll/tests/string.c
+++ b/dlls/ntdll/tests/string.c
@@ -68,7 +68,6 @@
 static WCHAR    (__cdecl *ptowupper)(WCHAR);
 static int      (__cdecl *p_wcsicmp)(LPCWSTR,LPCWSTR);
 static int      (__cdecl *p_wcsnicmp)(LPCWSTR,LPCWSTR,int);
-static LPWSTR   (__cdecl *pwcsncpy)(LPCWSTR,LPCWSTR,int);
 
 static LPWSTR   (__cdecl *pwcschr)(LPCWSTR, WCHAR);
 static LPWSTR   (__cdecl *pwcsrchr)(LPCWSTR, WCHAR);
@@ -140,7 +139,6 @@
     X(towupper);
     X(_wcsicmp);
     X(_wcsnicmp);
-    X(wcsncpy);
     X(wcschr);
     X(wcsrchr);
     X(memchr);
@@ -1691,19 +1689,19 @@
     memset(buffer, 0xcc, sizeof(buffer));
     res = p_snwprintf(buffer, 3, L"test");
     ok(res == -1, "res = %d\n", res);
-    ok(!memcmp(buffer, L"tes", 3 * sizeof(WCHAR)), "buf = %s\n", debugstr_wn(buffer, 3));
+    ok(!memcmp(buffer, L"tes", 3 * sizeof(WCHAR)), "buf = %s\n", debugstr_w(buffer));
     ok(buffer[3] == 0xcccc, "buffer[3] = %x\n", buffer[3]);
 
     memset(buffer, 0xcc, sizeof(buffer));
     res = p_snwprintf(buffer, 4, L"%s", L"test");
     ok(res == 4, "res = %d\n", res);
-    ok(!memcmp(buffer, L"test", 4 * sizeof(WCHAR)), "buf = %s\n", debugstr_wn(buffer, 4));
+    ok(!memcmp(buffer, L"test", 4 * sizeof(WCHAR)), "buf = %s\n", debugstr_w(buffer));
     ok(buffer[4] == 0xcccc, "buffer[4] = %x\n", buffer[4]);
 
     memset(buffer, 0xcc, sizeof(buffer));
     res = p_snwprintf(buffer, 3, L"%s", L"test");
     ok(res == -1, "res = %d\n", res);
-    ok(!memcmp(buffer, L"tes", 3), "buf = %s\n", debugstr_wn(buffer, 3));
+    ok(!memcmp(buffer, L"tes", 3), "buf = %s\n", debugstr_w(buffer));
     ok(buffer[3] == 0xcccc, "buffer[3] = %x\n", buffer[3]);
 
     res = p_snwprintf(buffer, ARRAY_SIZE(buffer), L"%I64x %d", (ULONGLONG)0x1234567890, 1);
@@ -1921,7 +1919,7 @@
     WCHAR ws[256], expectedw[256], specw[256];
     unsigned int i, j;
     char expected[256], spec[256], s[256];
-    int len = 0;
+    int len;
 
     p_snprintf(s, ARRAY_SIZE(s), "%C", 0x95c8);
     p_snwprintf(ws, ARRAY_SIZE(ws), L"%c", 0x95c8);
@@ -1965,6 +1963,7 @@
                     len = p_snwprintf(ws, ~(size_t)0, specw, (ULONG64)tests[i].arg, L"end");
                 break;
             default:
+                len = 0;
                 ok(0, "unknown length %u.\n", tests[i].arg_size);
                 break;
         }
@@ -2234,23 +2233,6 @@
     ok(r == s, "memchr returned %p, expected %p\n", r, s);
 }
 
-static void test_wcsncpy(void)
-{
-    wchar_t dst[8], *p;
-
-    memset(dst, 0xbb, sizeof(dst));
-    p = pwcsncpy(dst, L"1234567", 6);
-    ok(p == dst, "unexpected return value.\n");
-    ok(!wcsncmp(dst, L"123456", 6), "unexpected buffer %s\n",
-            wine_dbgstr_wn(dst, ARRAY_SIZE(dst)));
-    ok(dst[6] == 0xbbbb , "expected 0xbbbb unexpected wchar_t %x\n", dst[6]);
-
-    memset(dst, 0xbb, sizeof(dst));
-    pwcsncpy(dst, L"123", 5);
-    ok(dst[4] == 0, "expected 0, unexpected wchar_t %x\n", dst[4]);
-    ok(dst[5] == 0xbbbb , "expected 0xbbbb unexpected wchar_t %x\n", dst[5]);
-}
-
 START_TEST(string)
 {
     InitFunctionPtrs();
@@ -2284,5 +2266,4 @@
     test_wctype();
     test_ctype();
     test_memchr();
-    test_wcsncpy();
-}
+}
