diff --git a/dlls/wow64/virtual.c b/dlls/wow64/virtual.c
index cbaf6dd..8f88be7 100644
--- a/dlls/wow64/virtual.c
+++ b/dlls/wow64/virtual.c
@@ -229,36 +229,6 @@ NTSTATUS WINAPI wow64_NtAreMappedFilesTheSame( UINT *args )
 }
 
 
-/**********************************************************************
- *           wow64_NtCreateSectionEx
- */
-NTSTATUS WINAPI wow64_NtCreateSectionEx( UINT *args )
-{
-    ULONG *handle_ptr = get_ptr( &args );
-    ACCESS_MASK access = get_ulong( &args );
-    OBJECT_ATTRIBUTES32 *attr32 = get_ptr( &args );
-    const LARGE_INTEGER *size = get_ptr( &args );
-    ULONG protect = get_ulong( &args );
-    ULONG flags = get_ulong( &args );
-    HANDLE file = get_handle( &args );
-    MEM_EXTENDED_PARAMETER32 *params32 = get_ptr( &args );
-    ULONG count = get_ulong( &args );
-
-    MEM_EXTENDED_PARAMETER *params64;
-    struct object_attr64 attr;
-    HANDLE handle = 0;
-    NTSTATUS status;
-
-    if ((status = mem_extended_parameters_32to64( &params64, params32, &count, FALSE ))) return status;
-
-    *handle_ptr = 0;
-    status = NtCreateSectionEx( &handle, access, objattr_32to64( &attr, attr32 ),
-                                size, protect, flags, file, params64, count );
-    put_handle( handle_ptr, handle );
-    return status;
-}
-
-
 /**********************************************************************
  *           wow64_NtFlushInstructionCache
  */
@@ -779,11 +749,15 @@ NTSTATUS WINAPI wow64_NtSetInformationVirtualMemory( UINT *args )
 NTSTATUS WINAPI wow64_NtSetLdtEntries( UINT *args )
 {
     ULONG sel1 = get_ulong( &args );
-    ULONG64 entry1 = get_ulong64( &args );
+    ULONG entry1_low = get_ulong( &args );
+    ULONG entry1_high = get_ulong( &args );
     ULONG sel2 = get_ulong( &args );
-    ULONG64 entry2 = get_ulong64( &args );
+    ULONG entry2_low = get_ulong( &args );
+    ULONG entry2_high = get_ulong( &args );
 
-    return NtSetLdtEntries( sel1, *(LDT_ENTRY *)&entry1, sel2, *(LDT_ENTRY *)&entry2 );
+    FIXME( "%04lx %08lx %08lx %04lx %08lx %08lx: stub\n",
+           sel1, entry1_low, entry1_high, sel2, entry2_low, entry2_high );
+    return STATUS_NOT_IMPLEMENTED;
 }
 
 
