diff --git a/dlls/imm32/ime.c b/dlls/imm32/ime.c
--- a/dlls/imm32/ime.c
+++ b/dlls/imm32/ime.c
@@ -205,11 +205,7 @@
     WCHAR *str;
     UINT len;
 
-    if (!(ctx = ImmLockIMC( himc )))
-    {
-        ValidateRect( hwnd, NULL );
-        return;
-    }
+    if (!(ctx = ImmLockIMC( himc ))) return;
 
     hdc = BeginPaint( hwnd, &ps );
 
@@ -327,6 +323,8 @@
         RedrawWindow( hwnd, NULL, NULL, RDW_ERASENOW | RDW_INVALIDATE );
     }
     free( str );
+
+    ctx->hWnd = GetFocus();
 }
 
 static void ime_ui_composition( HIMC himc, HWND hwnd, LPARAM lparam )
@@ -435,7 +433,7 @@
     case WM_IME_CONTROL:
         FIXME( "hwnd %p, himc %p, msg %s, wparam %s, lparam %#Ix stub!\n", hwnd, himc,
                debugstr_wm_ime(msg), debugstr_imc(wparam), lparam );
-        return 1;
+        return 0;
     }
 
     return DefWindowProcW( hwnd, msg, wparam, lparam );
@@ -531,7 +529,6 @@
     UINT size, count = 0;
     INPUTCONTEXT *ctx;
     NTSTATUS status;
-    BOOL key_consumed = TRUE;
 
     TRACE( "vkey %#x, vsc %#x, state %p, msgs %p, flags %#x, himc %p\n",
            vkey, vsc, state, msgs, flags, himc );
@@ -550,7 +547,6 @@
         if (!(compstr = ImmLockIMCC( (ctx->hCompStr = himcc) ))) goto done;
 
         params.compstr = compstr;
-        params.key_consumed = &key_consumed;
         status = NtUserMessageCall( ctx->hWnd, WINE_IME_TO_ASCII_EX, vkey, vsc, &params,
                                     NtUserImeDriverCall, FALSE );
         size = compstr->dwSize;
@@ -559,17 +555,14 @@
     if (status) WARN( "WINE_IME_TO_ASCII_EX returned status %#lx\n", status );
     else
     {
-        if (compstr->dwCompStrOffset || compstr->dwResultStrLen)
-        {
-            TRANSMSG msg = {.message = ime_set_composition_status( himc, TRUE )};
-            if (msg.message == WM_IME_STARTCOMPOSITION) msgs->TransMsg[count++] = msg;
-        }
-
-        if (compstr->dwResultStrLen)
+        TRANSMSG status_msg = {.message = ime_set_composition_status( himc, !!compstr->dwCompStrOffset )};
+        if (status_msg.message) msgs->TransMsg[count++] = status_msg;
+
+        if (compstr->dwResultStrOffset)
         {
             const WCHAR *result = (WCHAR *)((BYTE *)compstr + compstr->dwResultStrOffset);
             TRANSMSG msg = {.message = WM_IME_COMPOSITION, .wParam = result[0], .lParam = GCS_RESULTSTR};
-            if (compstr->dwResultClauseLen) msg.lParam |= GCS_RESULTCLAUSE;
+            if (compstr->dwResultClauseOffset) msg.lParam |= GCS_RESULTCLAUSE;
             msgs->TransMsg[count++] = msg;
         }
 
@@ -580,18 +573,6 @@
             if (compstr->dwCompAttrOffset) msg.lParam |= GCS_COMPATTR;
             if (compstr->dwCompClauseOffset) msg.lParam |= GCS_COMPCLAUSE;
             else msg.lParam |= CS_INSERTCHAR|CS_NOMOVECARET;
-            msgs->TransMsg[count++] = msg;
-        }
-
-        if (!compstr->dwCompStrLen)
-        {
-            TRANSMSG msg = {.message = ime_set_composition_status( himc, FALSE )};
-            if (msg.message == WM_IME_ENDCOMPOSITION) msgs->TransMsg[count++] = msg;
-        }
-
-        if (!key_consumed)
-        {
-            TRANSMSG msg = {.message = WM_IME_KEYDOWN, .wParam = vkey, .lParam = vsc};
             msgs->TransMsg[count++] = msg;
         }
     }
