diff --git a/dlls/wintab32/wintab32.c b/dlls/wintab32/wintab32.c
--- a/dlls/wintab32/wintab32.c
+++ b/dlls/wintab32/wintab32.c
@@ -24,7 +24,7 @@
 #include "winbase.h"
 #include "winreg.h"
 #include "wingdi.h"
-#include "ntuser.h"
+#include "winuser.h"
 #include "winerror.h"
 #define NOFIX32
 #include "wintab.h"
@@ -41,6 +41,11 @@
       0, 0, { (DWORD_PTR)(__FILE__ ": csTablet") }
 };
 CRITICAL_SECTION csTablet = { &csTablet_debug, -1, 0, 0, 0, 0 };
+
+int  (CDECL *pLoadTabletInfo)(HWND hwnddefault) = NULL;
+int  (CDECL *pGetCurrentPacket)(LPWTPACKET packet) = NULL;
+int  (CDECL *pAttachEventQueueToTablet)(HWND hOwner) = NULL;
+UINT (CDECL *pWTInfoW)(UINT wCategory, UINT nIndex, LPVOID lpOutput) = NULL;
 
 static LRESULT WINAPI TABLET_WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam,
                                           LPARAM lParam);
@@ -64,6 +69,30 @@
     UnregisterClassW(L"WineTabletClass", NULL);
 }
 
+static HMODULE load_graphics_driver(void)
+{
+    static const WCHAR key_pathW[] = L"System\\CurrentControlSet\\Control\\Video\\{";
+    static const WCHAR displayW[] = L"}\\0000";
+
+    HMODULE ret = 0;
+    HKEY hkey;
+    DWORD size;
+    WCHAR path[MAX_PATH];
+    WCHAR key[ARRAY_SIZE(key_pathW) + ARRAY_SIZE(displayW) + 40];
+    UINT guid_atom = HandleToULong( GetPropW( GetDesktopWindow(), L"__wine_display_device_guid" ));
+
+    if (!guid_atom) return 0;
+    memcpy( key, key_pathW, sizeof(key_pathW) );
+    if (!GlobalGetAtomNameW( guid_atom, key + lstrlenW(key), 40 )) return 0;
+    lstrcatW( key, displayW );
+    if (RegOpenKeyW( HKEY_LOCAL_MACHINE, key, &hkey )) return 0;
+    size = sizeof(path);
+    if (!RegQueryValueExW( hkey, L"GraphicsDriver", NULL, NULL, (BYTE *)path, &size )) ret = LoadLibraryW( path );
+    RegCloseKey( hkey );
+    TRACE( "%s %p\n", debugstr_w(path), ret );
+    return ret;
+}
+
 BOOL WINAPI DllMain(HINSTANCE hInstDLL, DWORD fdwReason, LPVOID lpReserved)
 {
     TRACE("%p, %lx, %p\n",hInstDLL,fdwReason,lpReserved);
@@ -75,7 +104,15 @@
             TABLET_Register();
             hwndDefault = CreateWindowW(L"WineTabletClass", L"Tablet",
                                         WS_POPUPWINDOW,0,0,0,0,0,0,hInstDLL,0);
-            if (!hwndDefault)
+            if (hwndDefault)
+            {
+                HMODULE module = load_graphics_driver();
+                pLoadTabletInfo = (void *)GetProcAddress(module, "LoadTabletInfo");
+                pAttachEventQueueToTablet = (void *)GetProcAddress(module, "AttachEventQueueToTablet");
+                pGetCurrentPacket = (void *)GetProcAddress(module, "GetCurrentPacket");
+                pWTInfoW = (void *)GetProcAddress(module, "WTInfoW");
+            }
+            else
                 return FALSE;
             break;
         case DLL_PROCESS_DETACH:
@@ -108,8 +145,9 @@
             {
                 WTPACKET packet;
                 LPOPENCONTEXT handler;
-                if (NtUserMessageCall(hwnd, NtUserWintabPacket, 0, 0, &packet, NtUserWintabDriverCall, FALSE))
+                if (pGetCurrentPacket)
                 {
+                    pGetCurrentPacket(&packet);
                     handler = AddPacketToContextQueue(&packet,(HWND)lParam);
                     if (handler && handler->context.lcOptions & CXO_MESSAGES)
                        TABLET_PostTabletMessage(handler, _WT_PACKET(handler->context.lcMsgBase),
@@ -122,8 +160,9 @@
             {
                 WTPACKET packet;
                 LPOPENCONTEXT handler;
-                if (NtUserMessageCall(hwnd, NtUserWintabPacket, 0, 0, &packet, NtUserWintabDriverCall, FALSE))
+                if (pGetCurrentPacket)
                 {
+                    pGetCurrentPacket(&packet);
                     handler = AddPacketToContextQueue(&packet,(HWND)wParam);
                     if (handler)
                         TABLET_PostTabletMessage(handler, WT_PROXIMITY,
