diff --git a/dlls/fusion/asmcache.c b/dlls/fusion/asmcache.c
--- a/dlls/fusion/asmcache.c
+++ b/dlls/fusion/asmcache.c
@@ -53,6 +53,9 @@
     LONG ref;
 } IAssemblyCacheItemImpl;
 
+static const WCHAR cache_mutex_nameW[] =
+    {'_','_','W','I','N','E','_','F','U','S','I','O','N','_','C','A','C','H','E','_','M','U','T','E','X','_','_',0};
+
 static BOOL create_full_path(LPCWSTR path)
 {
     LPWSTR new_path;
@@ -103,8 +106,11 @@
 
 static BOOL get_assembly_directory(LPWSTR dir, DWORD size, const char *version, PEKIND architecture)
 {
-    static const WCHAR dotnet[] = L"\\Microsoft.NET\\";
-    static const WCHAR gac[] = L"\\assembly\\GAC";
+    static const WCHAR dotnet[] = {'\\','M','i','c','r','o','s','o','f','t','.','N','E','T','\\',0};
+    static const WCHAR gac[] = {'\\','a','s','s','e','m','b','l','y','\\','G','A','C',0};
+    static const WCHAR msil[] = {'_','M','S','I','L',0};
+    static const WCHAR x86[] = {'_','3','2',0};
+    static const WCHAR amd64[] = {'_','6','4',0};
     DWORD len = GetWindowsDirectoryW(dir, size);
 
     if (!strcmp(version, "v4.0.30319"))
@@ -125,15 +131,15 @@
             break;
 
         case peMSIL:
-            lstrcpyW(dir + len, L"_MSIL");
+            lstrcpyW(dir + len, msil);
             break;
 
         case peI386:
-            lstrcpyW(dir + len, L"_32");
+            lstrcpyW(dir + len, x86);
             break;
 
         case peAMD64:
-            lstrcpyW(dir + len, L"_64");
+            lstrcpyW(dir + len, amd64);
             break;
 
         default:
@@ -413,6 +419,12 @@
                                                          LPCWSTR pszManifestFilePath,
                                                          LPCFUSION_INSTALL_REFERENCE pRefData)
 {
+    static const WCHAR format[] =
+        {'%','s','\\','%','s','\\','%','s','_','_','%','s','\\',0};
+    static const WCHAR format_v40[] =
+        {'%','s','\\','%','s','\\','v','4','.','0','_','%','s','_','_','%','s','\\',0};
+    static const WCHAR ext_exe[] = {'.','e','x','e',0};
+    static const WCHAR ext_dll[] = {'.','d','l','l',0};
     IAssemblyCacheImpl *cache = impl_from_IAssemblyCache(iface);
     ASSEMBLY *assembly;
     const WCHAR *extension, *filename, *src_dir;
@@ -420,7 +432,7 @@
     WCHAR asmdir[MAX_PATH], *p, **external_files = NULL, *dst_dir = NULL;
     PEKIND architecture;
     char *clr_version;
-    DWORD i, count = 0, src_len, dst_len = sizeof("\\\\v4.0___\\");
+    DWORD i, count = 0, src_len, dst_len = ARRAY_SIZE(format_v40);
     HRESULT hr;
 
     TRACE("(%p, %ld, %s, %p)\n", iface, dwFlags,
@@ -432,7 +444,7 @@
     if (!(extension = wcsrchr(pszManifestFilePath, '.')))
         return HRESULT_FROM_WIN32(ERROR_INVALID_NAME);
 
-    if (lstrcmpiW(extension, L".exe") && lstrcmpiW(extension, L".dll"))
+    if (lstrcmpiW(extension, ext_exe) && lstrcmpiW(extension, ext_dll))
         return HRESULT_FROM_WIN32(ERROR_INVALID_NAME);
 
     if (GetFileAttributesW(pszManifestFilePath) == INVALID_FILE_ATTRIBUTES)
@@ -477,9 +489,9 @@
         goto done;
     }
     if (!strcmp(clr_version, "v4.0.30319"))
-        dst_len = swprintf(dst_dir, dst_len, L"%s\\%s\\v4.0_%s__%s\\", asmdir, name, version, token);
+        dst_len = swprintf(dst_dir, dst_len, format_v40, asmdir, name, version, token);
     else
-        dst_len = swprintf(dst_dir, dst_len, L"%s\\%s\\%s__%s\\", asmdir, name, version, token);
+        dst_len = swprintf(dst_dir, dst_len, format, asmdir, name, version, token);
 
     create_full_path(dst_dir);
 
@@ -552,7 +564,7 @@
 
     cache->IAssemblyCache_iface.lpVtbl = &AssemblyCacheVtbl;
     cache->ref = 1;
-    cache->lock = CreateMutexW( NULL, FALSE, L"__WINE_FUSION_CACHE_MUTEX__" );
+    cache->lock = CreateMutexW( NULL, FALSE, cache_mutex_nameW );
     if (!cache->lock)
     {
         free( cache );
