diff --git a/dlls/ntdll/tests/wow64.c b/dlls/ntdll/tests/wow64.c
--- a/dlls/ntdll/tests/wow64.c
+++ b/dlls/ntdll/tests/wow64.c
@@ -42,7 +42,6 @@
 static NTSTATUS (WINAPI *pRtlWow64GetThreadContext)(HANDLE,WOW64_CONTEXT*);
 static NTSTATUS (WINAPI *pRtlWow64IsWowGuestMachineSupported)(USHORT,BOOLEAN*);
 static NTSTATUS (WINAPI *pNtMapViewOfSectionEx)(HANDLE,HANDLE,PVOID*,const LARGE_INTEGER*,SIZE_T*,ULONG,ULONG,MEM_EXTENDED_PARAMETER*,ULONG);
-static NTSTATUS (WINAPI *pNtSetLdtEntries)(ULONG,LDT_ENTRY,ULONG,LDT_ENTRY);
 #ifdef _WIN64
 static NTSTATUS (WINAPI *pKiUserExceptionDispatcher)(EXCEPTION_RECORD*,CONTEXT*);
 static NTSTATUS (WINAPI *pRtlWow64GetCpuAreaInfo)(WOW64_CPURESERVED*,ULONG,WOW64_CPU_AREA_INFO*);
@@ -118,7 +117,6 @@
 
 #define GET_PROC(func) p##func = (void *)GetProcAddress( ntdll, #func )
     GET_PROC( NtMapViewOfSectionEx );
-    GET_PROC( NtSetLdtEntries );
     GET_PROC( NtQuerySystemInformation );
     GET_PROC( NtQuerySystemInformationEx );
     GET_PROC( RtlGetNativeSystemInformation );
@@ -1174,7 +1172,6 @@
     THREAD_DESCRIPTOR_INFORMATION info;
     NTSTATUS status;
     ULONG base, limit, sel, retlen;
-    LDT_ENTRY ds_entry = { 0 };
     I386_CONTEXT context = { CONTEXT_I386_CONTROL | CONTEXT_I386_SEGMENTS };
 
 #ifdef _WIN64
@@ -1273,7 +1270,6 @@
             ok( !info.Entry.HighWord.Bits.Sys, "wrong sys\n" );
             ok( info.Entry.HighWord.Bits.Default_Big, "wrong big\n" );
             ok( info.Entry.HighWord.Bits.Granularity, "wrong granularity\n" );
-            ds_entry = info.Entry;
         }
         else if (sel == context.SegFs)  /* TEB selector */
         {
@@ -1298,9 +1294,8 @@
         else if (!status)
         {
             ok( retlen == sizeof(info.Entry), "len set %lu\n", retlen );
-            trace( "succeeded for %04lx base %lx limit %lx type %x\n",
+            trace( "succeeded for %lx base %lx limit %lx type %x\n",
                    sel, base, limit, info.Entry.HighWord.Bits.Type );
-            ok( !(sel & 4), "succeeded for LDT selector %04lx\n", sel );
         }
         else
         {
@@ -1311,29 +1306,6 @@
             ok( retlen == 0xdeadbeef, "len set %lu\n", retlen );
         }
     }
-
-    status = pNtSetLdtEntries( 0, ds_entry, 0, ds_entry );
-    if (status != STATUS_NOT_IMPLEMENTED)
-    {
-        ok( !status, "NtSetLdtEntries failed: %08lx\n", status );
-
-        status = pNtSetLdtEntries( 0x000f, ds_entry, 0x001f, ds_entry );
-        ok( !status, "NtSetLdtEntries failed: %08lx\n", status );
-
-        info.Selector = 0x000f;
-        memset(&info.Entry, 0x9a, sizeof(info.Entry));
-        status = GET_ENTRY( &info, sizeof(info), NULL );
-        ok(!status, "wrong status %lx\n", status);
-        ok(!memcmp(&ds_entry, &info.Entry, sizeof(ds_entry)), "entries do not match\n");
-
-        info.Selector = 0x001f;
-        memset(&info.Entry, 0x9a, sizeof(info.Entry));
-        status = GET_ENTRY( &info, sizeof(info), NULL );
-        ok(!status, "wrong status %lx\n", status);
-        ok(!memcmp(&ds_entry, &info.Entry, sizeof(ds_entry)), "entries do not match\n");
-    }
-    else skip( "NtSetLdtEntries not supported\n" );
-
 #undef GET_ENTRY
 #endif /* __arm__ */
 }
@@ -2520,10 +2492,10 @@
                                                       MEM_RESERVE | MEM_COMMIT, PAGE_READONLY );
             ok( !status, "NtWow64AllocateVirtualMemory64 failed %lx\n", status );
             status = pNtWow64WriteVirtualMemory64( process, ptr, str, sizeof(str), &res );
-            todo_wine_if(status == STATUS_SUCCESS)
+            todo_wine
             ok( status == STATUS_PARTIAL_COPY || broken( status == STATUS_ACCESS_VIOLATION ),
                 "NtWow64WriteVirtualMemory64 failed %lx\n", status );
-            todo_wine_if(status == STATUS_SUCCESS)
+            todo_wine
             ok( !res || broken(res) /* win10 1709 */, "wrong size %s\n", wine_dbgstr_longlong(res) );
         }
         ptr = 0x9876543210ull;
@@ -2615,7 +2587,7 @@
 
         if (native_machine == IMAGE_FILE_MACHINE_ARM64)
         {
-            KUSER_SHARED_DATA *user_shared_data = ULongToPtr( 0x7ffe0000 );
+            KSHARED_USER_DATA *user_shared_data = ULongToPtr( 0x7ffe0000 );
 
             ok( user_shared_data->ProcessorFeatures[PF_ARM_V8_INSTRUCTIONS_AVAILABLE], "no ARM_V8\n" );
             ok( user_shared_data->ProcessorFeatures[PF_MMX_INSTRUCTIONS_AVAILABLE], "no MMX\n" );
