diff --git a/dlls/comctl32/tests/trackbar.c b/dlls/comctl32/tests/trackbar.c
--- a/dlls/comctl32/tests/trackbar.c
+++ b/dlls/comctl32/tests/trackbar.c
@@ -50,53 +50,7 @@
 
 static struct msg_sequence *sequences[NUM_MSG_SEQUENCE];
 
-static void CALLBACK msg_winevent_proc(HWINEVENTHOOK hevent,
-                                       DWORD event,
-                                       HWND hwnd,
-                                       LONG object_id,
-                                       LONG child_id,
-                                       DWORD thread_id,
-                                       DWORD event_time)
-{
-    struct message msg = {0};
-    char class_name[256];
-
-    /* ignore window and other system events */
-    if (object_id != OBJID_CLIENT) return;
-
-    /* ignore events not from a progress bar control */
-    if (!GetClassNameA(hwnd, class_name, ARRAY_SIZE(class_name)) ||
-        strcmp(class_name, TRACKBAR_CLASSA) != 0)
-        return;
-
-    msg.message = event;
-    msg.flags = winevent_hook|wparam|lparam;
-    msg.wParam = object_id;
-    msg.lParam = child_id;
-    add_message(sequences, TRACKBAR_SEQ_INDEX, &msg);
-}
-
-static void init_winevent_hook(void) {
-    hwineventhook = SetWinEventHook(EVENT_MIN, EVENT_MAX, GetModuleHandleA(0), msg_winevent_proc,
-                                    0, GetCurrentThreadId(), WINEVENT_INCONTEXT);
-    if (!hwineventhook)
-        win_skip( "no win event hook support\n" );
-}
-
-static void uninit_winevent_hook(void) {
-    if (!hwineventhook)
-        return;
-
-    UnhookWinEvent(hwineventhook);
-    hwineventhook = 0;
-}
-
 static const struct message empty_seq[] = {
-    {0}
-};
-
-static const struct message create_trackbar_wnd_seq[] = {
-    { EVENT_OBJECT_FOCUS, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     {0}
 };
 
@@ -176,14 +130,10 @@
     { TBM_GETPOS, sent},
     { TBM_SETPOS, sent|wparam|lparam, TRUE, 5},
     { WM_PAINT, sent|defwinproc},
-    { EVENT_OBJECT_STATECHANGE, winevent_hook|wparam|lparam|optional, OBJID_CLIENT, 3 }, /* v6 only */
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { TBM_GETPOS, sent},
     { TBM_SETPOS, sent|wparam|lparam, TRUE, 5},
     { TBM_SETPOS, sent|wparam|lparam, TRUE, 1000},
     { WM_PAINT, sent|defwinproc},
-    { EVENT_OBJECT_STATECHANGE, winevent_hook|wparam|lparam|optional, OBJID_CLIENT, 3 }, /* v6 only */
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { TBM_GETPOS, sent},
     { TBM_SETPOS, sent|wparam|lparam, FALSE, 20},
     { TBM_GETPOS, sent},
@@ -1405,7 +1355,7 @@
 
     hWndTrackbar = create_trackbar(defaultstyle, hWndParent);
     ok(hWndTrackbar != NULL, "Expected non NULL value\n");
-    ok_sequence(sequences, TRACKBAR_SEQ_INDEX, create_trackbar_wnd_seq, "create Trackbar Window", FALSE);
+    ok_sequence(sequences, TRACKBAR_SEQ_INDEX, empty_seq, "create Trackbar Window", FALSE);
     ok_sequence(sequences, PARENT_SEQ_INDEX, parent_create_trackbar_wnd_seq, "parent trackbar window", TRUE);
 
     DestroyWindow(hWndTrackbar);
@@ -1435,8 +1385,6 @@
         skip("parent window not present\n");
         return;
     }
-
-    init_winevent_hook();
 
     test_create();
     test_trackbar_buddy();
@@ -1472,8 +1420,6 @@
     test_ignore_selection();
     test_initial_state();
 
-    uninit_winevent_hook();
-
     unload_v6_module(cookie, ctxt);
 
     DestroyWindow(hWndParent);
