diff --git a/dlls/taskschd/regtask.c b/dlls/taskschd/regtask.c
--- a/dlls/taskschd/regtask.c
+++ b/dlls/taskschd/regtask.c
@@ -464,90 +464,15 @@
 
 static HRESULT WINAPI regtasks_get_Count(IRegisteredTaskCollection *iface, LONG *count)
 {
-    HRESULT hr;
-    RegisteredTaskCollection *reg_tasks = NULL;
-    TASK_NAMES task_names = NULL;
-    DWORD start_index = 0, num_tasks;
-
-    reg_tasks = impl_from_IRegisteredTaskCollection(iface);
-
-    if (!count)
-        return E_POINTER;
-
-    hr = SchRpcEnumTasks(reg_tasks->path, 0, &start_index, 0, &num_tasks, &task_names);
-    if (FAILED(hr))
-    {
-        *count = 0;
-        return hr;
-    }
-
-    *count = num_tasks;
-
-    for (DWORD i = 0; i < num_tasks; i++)
-        MIDL_user_free(task_names[i]);
-    MIDL_user_free(task_names);
-
-    return S_OK;
+    FIXME("%p,%p: stub\n", iface, count);
+    if (count) *count = 0;
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI regtasks_get_Item(IRegisteredTaskCollection *iface, VARIANT index, IRegisteredTask **regtask)
 {
-    HRESULT hr;
-    VARIANT converted_index;
-    TASK_NAMES task_names = NULL;
-    ITaskDefinition *definition = NULL;
-    RegisteredTaskCollection *collection = NULL;
-    DWORD start_index = 0, num_tasks = 0;
-
-    collection = impl_from_IRegisteredTaskCollection(iface);
-    if (!regtask)
-        return E_POINTER;
-    *regtask = NULL;
-
-    VariantInit(&converted_index);
-    hr = VariantChangeType(&converted_index, &index, 0, VT_UI4);
-    if (FAILED(hr))
-        return hr;
-
-    start_index = V_UI4(&index);
-    if (start_index == 0)
-    {
-        VariantClear(&converted_index);
-        return E_INVALIDARG;
-    }
-    start_index -= 1;
-
-    hr = SchRpcEnumTasks(collection->path, 0, &start_index, 1, &num_tasks, &task_names);
-    if (FAILED(hr))
-    {
-        VariantClear(&converted_index);
-        return hr;
-    }
-    if (!task_names)
-    {
-        VariantClear(&converted_index);
-        return E_INVALIDARG;
-    }
-
-    hr = TaskDefinition_create(&definition);
-    if (FAILED(hr))
-    {
-        if(hr != E_OUTOFMEMORY)
-            ITaskDefinition_Release(definition);
-        VariantClear(&converted_index);
-        return hr;
-    }
-
-    hr = RegisteredTask_create(collection->path, task_names[0], definition, TASK_VALIDATE_ONLY, TASK_LOGON_INTERACTIVE_TOKEN, regtask, FALSE);
-
-    VariantClear(&converted_index);
-    MIDL_user_free(task_names[0]);
-    MIDL_user_free(task_names);
-
-    if (hr != S_OK)
-        ITaskDefinition_Release(definition);
-
-    return hr;
+    FIXME("%p,%s,%p: stub\n", iface, debugstr_variant(&index), regtask);
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI regtasks_get__NewEnum(IRegisteredTaskCollection *iface, IUnknown **penum)
