diff --git a/dlls/kernel32/tests/heap.c b/dlls/kernel32/tests/heap.c
--- a/dlls/kernel32/tests/heap.c
+++ b/dlls/kernel32/tests/heap.c
@@ -3404,7 +3404,12 @@
     sprintf( buffer, "%s heap.c 0x%lx", argv0, flags );
     ret = CreateProcessA( NULL, buffer, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &info );
     ok( ret, "failed to create child process error %lu\n", GetLastError() );
-    wait_child_process( &info );
+    if (ret)
+    {
+        wait_child_process( info.hProcess );
+        CloseHandle( info.hThread );
+        CloseHandle( info.hProcess );
+    }
     RegDeleteValueA( hkey, "GlobalFlag" );
     RegCloseKey( hkey );
     RegDeleteKeyA( HKEY_LOCAL_MACHINE, keyname );
@@ -3664,9 +3669,9 @@
 
     ok( memex.dwMemoryLoad == expect.dwMemoryLoad, "got dwMemoryLoad %lu\n", memex.dwMemoryLoad );
     ok( memex.ullTotalPhys == expect.ullTotalPhys, "got ullTotalPhys %#I64x\n", memex.ullTotalPhys );
-    flaky ok( IS_WITHIN_RANGE( memex.ullAvailPhys, expect.ullAvailPhys ), "got ullAvailPhys %#I64x\n", memex.ullAvailPhys );
+    ok( IS_WITHIN_RANGE( memex.ullAvailPhys, expect.ullAvailPhys ), "got ullAvailPhys %#I64x\n", memex.ullAvailPhys );
     ok( memex.ullTotalPageFile == expect.ullTotalPageFile, "got ullTotalPageFile %#I64x\n", memex.ullTotalPageFile );
-    flaky ok( IS_WITHIN_RANGE( memex.ullAvailPageFile, expect.ullAvailPageFile ), "got ullAvailPageFile %#I64x\n", memex.ullAvailPageFile );
+    ok( IS_WITHIN_RANGE( memex.ullAvailPageFile, expect.ullAvailPageFile ), "got ullAvailPageFile %#I64x\n", memex.ullAvailPageFile );
     ok( memex.ullTotalVirtual == expect.ullTotalVirtual, "got ullTotalVirtual %#I64x\n", memex.ullTotalVirtual );
     ok( memex.ullAvailVirtual <= expect.ullAvailVirtual, "got ullAvailVirtual %#I64x\n", memex.ullAvailVirtual );
     ok( memex.ullAvailExtendedVirtual == 0, "got ullAvailExtendedVirtual %#I64x\n", memex.ullAvailExtendedVirtual );
@@ -3704,7 +3709,7 @@
     while (1)
     {
         size = VirtualQuery( p, &info2, sizeof(info2) );
-        if (!size) break;
+        ok( size == sizeof(info), "got %Iu.\n", size );
         if (info2.AllocationBase != info.AllocationBase)
             break;
         ok( info2.State == MEM_RESERVE || info2.State == MEM_COMMIT, "got %#lx.\n", info2.State );
@@ -3817,53 +3822,6 @@
     }
 }
 
-static void test_HeapSummary(void)
-{
-    HANDLE heap;
-    HEAP_SUMMARY heap_summary;
-    BOOL ret;
-    DWORD err;
-    void *p;
-
-    /* setup */
-
-    heap = HeapCreate( 0, 0, 0 ); /* growable heap */
-    ok( heap != NULL, "creation failed\n" );
-
-    HeapAlloc( heap , 0, 0x100 );
-    HeapAlloc( heap , 0, 0x200 );
-    p = HeapAlloc( heap , 0, 0x300 );
-    HeapAlloc( heap, 0, 0x60000 );
-    HeapFree( heap, 0, p );
-
-    memset( &heap_summary, 0, sizeof(heap_summary) );
-
-    /* test cases */
-
-    ret = HeapSummary( heap, 0, &heap_summary );
-    err = GetLastError();
-    ok( !ret, "HeapSummary() with cb != sizeof(HEAP_SUMMARY) returned TRUE\n" );
-    ok( err == ERROR_INVALID_PARAMETER,
-        "HeapSummary() with cb != sizeof(HEAP_SUMMARY) set last error to %lu\n", err );
-
-    heap_summary.cb = sizeof(heap_summary);
-    ret = HeapSummary( heap, 0, &heap_summary );
-    ok( ret, "HeapSummary() returned FALSE\n" );
-
-    ok( heap_summary.cbAllocated == 0x100 + 0x200 + 0x60000,
-        "HeapSummary: wrong cbAllocated value %#Ix\n", heap_summary.cbAllocated );
-    ok( heap_summary.cbCommitted >= heap_summary.cbAllocated,
-        "HeapSummary: cbCommitted %#Ix < cbAllocated %#Ix\n",
-        heap_summary.cbCommitted, heap_summary.cbAllocated );
-    ok( heap_summary.cbReserved >= heap_summary.cbCommitted,
-        "HeapSummary: cbReserved %#Ix < cbCommitted %#Ix\n",
-        heap_summary.cbReserved, heap_summary.cbCommitted );
-
-    /* cleanup */
-
-    HeapDestroy( heap );
-}
-
 START_TEST(heap)
 {
     int argc;
@@ -3884,7 +3842,6 @@
 
     test_GetPhysicallyInstalledSystemMemory();
     test_GlobalMemoryStatus();
-    test_HeapSummary();
     test_heap_tail_zeroing( 0 );
 
     if (pRtlGetNtGlobalFlags)
