diff --git a/dlls/user32/tests/cursoricon.c b/dlls/user32/tests/cursoricon.c
--- a/dlls/user32/tests/cursoricon.c
+++ b/dlls/user32/tests/cursoricon.c
@@ -31,7 +31,7 @@
 #include "wingdi.h"
 #include "winuser.h"
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 
 typedef struct
 {
@@ -291,12 +291,13 @@
     }
 };
 
-#pragma pack(pop)
+#include "poppack.h"
 
 static char **test_argv;
 static int test_argc;
 static HWND child = 0;
 static HWND parent = 0;
+static HANDLE child_process;
 
 #define PROC_INIT (WM_USER+1)
 
@@ -393,19 +394,14 @@
     }
 }
 
-static void test_child_process(void)
-{
-    static const BYTE bmp_bits[4096];
+static void do_parent(void)
+{
     char path_name[MAX_PATH];
     PROCESS_INFORMATION info;
     STARTUPINFOA startup;
-    ICONINFO cursorInfo;
-    UINT display_bpp;
     WNDCLASSA class;
-    HCURSOR cursor;
+    MSG msg;
     BOOL ret;
-    HDC hdc;
-    MSG msg;
 
     /* Register a new class. */
     class.style = CS_GLOBALCLASS;
@@ -436,6 +432,7 @@
 
     sprintf(path_name, "%s cursoricon %Ix", test_argv[0], (INT_PTR)parent);
     ok(CreateProcessA(NULL, path_name, NULL, NULL, FALSE, 0L, NULL, NULL, &startup, &info), "CreateProcess failed.\n");
+    child_process = info.hProcess;
 
     /* Wait for child window handle. */
     while ((child == 0) && (ret = GetMessageA(&msg, parent, 0, 0)))
@@ -444,6 +441,22 @@
         TranslateMessage(&msg);
         DispatchMessageA(&msg);
     }
+}
+
+static void finish_child_process(void)
+{
+    SendMessageA(child, WM_CLOSE, 0, 0);
+    wait_child_process( child_process );
+    CloseHandle(child_process);
+}
+
+static void test_child_process(void)
+{
+    static const BYTE bmp_bits[4096];
+    HCURSOR cursor;
+    ICONINFO cursorInfo;
+    UINT display_bpp;
+    HDC hdc;
 
     /* Create and set a dummy cursor. */
     hdc = GetDC(0);
@@ -463,9 +476,6 @@
 
     /* Destroy the cursor. */
     SendMessageA(child, WM_USER+1, 0, (LPARAM) cursor);
-
-    SendMessageA(child, WM_CLOSE, 0, 0);
-    wait_child_process( &info );
 }
 
 static BOOL color_match(COLORREF a, COLORREF b)
@@ -3149,5 +3159,7 @@
     test_DestroyCursor();
     test_PrivateExtractIcons();
     test_monochrome_icon();
+    do_parent();
     test_child_process();
-}
+    finish_child_process();
+}
