diff --git a/dlls/user32/listbox.c b/dlls/user32/listbox.c
--- a/dlls/user32/listbox.c
+++ b/dlls/user32/listbox.c
@@ -31,6 +31,7 @@
 #include "controls.h"
 #include "wine/exception.h"
 #include "wine/debug.h"
+#include "wine/heap.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(listbox);
 
@@ -148,7 +149,7 @@
         items_size = (items_size + LB_ARRAY_GRANULARITY - 1) & ~(LB_ARRAY_GRANULARITY - 1);
         if ((descr->style & (LBS_NODATA | LBS_MULTIPLESEL | LBS_EXTENDEDSEL)) != LBS_NODATA)
         {
-            items = realloc(descr->u.items, items_size * get_sizeof_item(descr));
+            items = heap_realloc(descr->u.items, items_size * get_sizeof_item(descr));
             if (!items)
             {
                 SEND_NOTIFICATION(descr, LBN_ERRSPACE);
@@ -457,31 +458,18 @@
     descr->height = rect.bottom - rect.top;
     if (!(descr->style & LBS_NOINTEGRALHEIGHT) && !(descr->style & LBS_OWNERDRAWVARIABLE))
     {
-        int height = descr->height;
         INT remaining;
         RECT rect;
 
-        /* The whole point of integral height is to ensure that partial items
-         * aren't displayed. Native seems to fail to take the horizontal
-         * scrollbar into account (while successfully taking into account e.g.
-         * WS_EX_CLIENTEDGE), so it ends up obscuring partial items anyway.
-         *
-         * It's not clear if native is trying to work from
-         * the window rect as opposed to the client rect [and badly
-         * reimplementing AdjustWindowRect()] or poorly working around the case
-         * where the horizontal scrollbar is repeatedly toggled (which could
-         * unnecessarily shrink the scrollbar every time it happens). */
-        if (GetWindowLongW( descr->self, GWL_STYLE ) & WS_HSCROLL)
-            height += GetSystemMetrics( SM_CYHSCROLL );
-
         GetWindowRect( descr->self, &rect );
         if(descr->item_height != 0)
-            remaining = height % descr->item_height;
+            remaining = descr->height % descr->item_height;
         else
             remaining = 0;
-        if ((height > descr->item_height) && remaining)
-        {
-            TRACE( "[%p]: changing height %d -> %d\n", descr->self, height, height - remaining );
+        if ((descr->height > descr->item_height) && remaining)
+        {
+            TRACE("[%p]: changing height %d -> %d\n",
+                  descr->self, descr->height, descr->height - remaining );
             NtUserSetWindowPos( descr->self, 0, 0, 0, rect.right - rect.left,
                                 rect.bottom - rect.top - remaining,
                                 SWP_NOZORDER | SWP_NOACTIVATE | SWP_NOMOVE );
@@ -620,7 +608,7 @@
     if (index < descr->nb_items)
     {
         item_str = get_item_string(descr, index);
-        selected = !(descr->style & LBS_NOSEL) && is_item_selected(descr, index);
+        selected = is_item_selected(descr, index);
     }
 
     focused = !ignoreFocus && descr->focus_item == index && descr->caret_on && descr->in_focus;
@@ -1801,7 +1789,7 @@
 
     if (!(descr->style & LBS_NODATA))
         for (i = descr->nb_items - 1; i >= 0; i--) LISTBOX_DeleteItem(descr, i);
-    free( descr->u.items );
+    HeapFree( GetProcessHeap(), 0, descr->u.items );
     descr->nb_items      = 0;
     descr->top_item      = 0;
     descr->selected_item = -1;
@@ -2297,7 +2285,7 @@
 static LRESULT LISTBOX_HandleLButtonUp( LB_DESCR *descr )
 {
     if (LISTBOX_Timer != LB_TIMER_NONE)
-        NtUserKillSystemTimer( descr->self, LB_TIMER_ID );
+        KillSystemTimer( descr->self, LB_TIMER_ID );
     LISTBOX_Timer = LB_TIMER_NONE;
     if (descr->captured)
     {
@@ -2354,7 +2342,7 @@
 {
     if (!LISTBOX_HandleTimer( descr, descr->focus_item, LISTBOX_Timer ))
     {
-        NtUserKillSystemTimer( descr->self, LB_TIMER_ID );
+        KillSystemTimer( descr->self, LB_TIMER_ID );
         LISTBOX_Timer = LB_TIMER_NONE;
     }
     return 0;
@@ -2406,7 +2394,7 @@
     if (dir != LB_TIMER_NONE)
         NtUserSetSystemTimer( descr->self, LB_TIMER_ID, LB_SCROLL_TIMEOUT );
     else if (LISTBOX_Timer != LB_TIMER_NONE)
-        NtUserKillSystemTimer( descr->self, LB_TIMER_ID );
+        KillSystemTimer( descr->self, LB_TIMER_ID );
     LISTBOX_Timer = dir;
 }
 
