diff --git a/dlls/taskschd/task.c b/dlls/taskschd/task.c
--- a/dlls/taskschd/task.c
+++ b/dlls/taskschd/task.c
@@ -2373,7 +2373,6 @@
     ITriggerCollection *triggers;
     IPrincipal *principal;
     IActionCollection *actions;
-    BSTR data;
 } TaskDefinition;
 
 static inline TaskDefinition *impl_from_ITaskDefinition(ITaskDefinition *iface)
@@ -2406,8 +2405,6 @@
             IPrincipal_Release(taskdef->principal);
         if (taskdef->actions)
             IActionCollection_Release(taskdef->actions);
-        if (taskdef->data)
-            SysFreeString(taskdef->data);
 
         free(taskdef);
     }
@@ -2579,41 +2576,14 @@
 
 static HRESULT WINAPI TaskDefinition_get_Data(ITaskDefinition *iface, BSTR *data)
 {
-    TaskDefinition *taskdef = impl_from_ITaskDefinition(iface);
-
-    TRACE("%p,%p\n", iface, data);
-
-    if (!data) return E_POINTER;
-
-    if (!taskdef->data)
-        *data = NULL;
-    else
-    {
-        *data = SysAllocString(taskdef->data);
-        if (!*data) return E_OUTOFMEMORY;
-    }
-
-    return S_OK;
+    FIXME("%p,%p: stub\n", iface, data);
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI TaskDefinition_put_Data(ITaskDefinition *iface, BSTR data)
 {
-    TaskDefinition *taskdef = impl_from_ITaskDefinition(iface);
-    BSTR copy = NULL;
-
-    TRACE("%p,%s\n", iface, debugstr_w(data));
-
-    if (data)
-    {
-        copy = SysAllocString(data);
-        if (!copy) return E_OUTOFMEMORY;
-    }
-
-    if (taskdef->data)
-        SysFreeString(taskdef->data);
-
-    taskdef->data = copy;
-    return S_OK;
+    FIXME("%p,%p: stub\n", iface, data);
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI TaskDefinition_get_Principal(ITaskDefinition *iface, IPrincipal **principal)
@@ -3156,12 +3126,6 @@
 
     hr = write_actions(stream, taskdef->actions);
     if (hr != S_OK) goto failed;
-
-    if (taskdef->data)
-    {
-        hr = write_text_value(stream, L"Data", taskdef->data);
-        if (hr != S_OK) goto failed;
-    }
 
     pop_indent();
 
@@ -3802,13 +3766,6 @@
                 hr = read_principals(reader, taskdef);
             else if (!lstrcmpW(name, L"Actions"))
                 hr = read_actions(reader, taskdef);
-            else if (!lstrcmpW(name, L"Data"))
-            {
-                WCHAR *value;
-                hr = read_text_value(reader, &value);
-                if (hr == S_OK)
-                    ITaskDefinition_put_Data(taskdef, value);
-            }
             else
                 FIXME("unhandled Task element %s\n", debugstr_w(name));
 
@@ -3937,11 +3894,6 @@
             IActionCollection_Release(taskdef->actions);
             taskdef->actions = NULL;
         }
-        if (taskdef->data)
-        {
-            SysFreeString(taskdef->data);
-            taskdef->data = NULL;
-        }
 
         hr = read_xml(reader, iface);
     }
