diff --git a/dlls/ws2_32/protocol.c b/dlls/ws2_32/protocol.c
--- a/dlls/ws2_32/protocol.c
+++ b/dlls/ws2_32/protocol.c
@@ -148,6 +148,25 @@
     return 0;
 }
 
+static BOOL eac_download_hack(void)
+{
+    static int eac_download_hack_enabled = -1;
+    char str[64];
+
+    if (eac_download_hack_enabled == -1)
+    {
+        if (GetEnvironmentVariableA("WINE_DISABLE_EAC_ALT_DOWNLOAD", str, sizeof(str)))
+            eac_download_hack_enabled = !!atoi(str);
+        else
+            eac_download_hack_enabled = GetEnvironmentVariableA("SteamGameId", str, sizeof(str))
+                                        && !strcmp(str, "626690");
+
+        if (eac_download_hack_enabled)
+            ERR("HACK: failing download-alt.easyanticheat.net resolution.\n");
+    }
+    return eac_download_hack_enabled;
+}
+
 /***********************************************************************
  *      getaddrinfo   (ws2_32.@)
  */
@@ -169,6 +188,12 @@
 
     if (node)
     {
+        if (eac_download_hack() && !strcmp(node, "download-alt.easyanticheat.net"))
+        {
+            SetLastError(WSAHOST_NOT_FOUND);
+            return WSAHOST_NOT_FOUND;
+        }
+
         if (!node[0])
         {
             if (!(fqdn = get_fqdn())) return WSA_NOT_ENOUGH_MEMORY;
@@ -927,6 +952,12 @@
         return NULL;
     }
 
+    if (eac_download_hack() && name && !strcmp(name, "download-alt.easyanticheat.net"))
+    {
+        SetLastError( WSAHOST_NOT_FOUND );
+        return NULL;
+    }
+
     if ((ret = WS_CALL( gethostname, &params )))
     {
         SetLastError( ret );
@@ -1794,20 +1825,7 @@
             sprintf( buffer + strlen( buffer ), "]:%u", ntohs( addr6->sin6_port ) );
         break;
     }
-    case AF_BTH:
-    {
-        const SOCKADDR_BTH *sockaddr_bth = (const SOCKADDR_BTH *)addr;
-        BLUETOOTH_ADDRESS addr_bth;
-
-        if (addr_len < sizeof(SOCKADDR_BTH)) return -1;
-
-        addr_bth.ullLong = sockaddr_bth->btAddr;
-        sprintf( buffer, "(%02X:%02X:%02X:%02X:%02X:%02X)", addr_bth.rgBytes[5], addr_bth.rgBytes[4],
-                 addr_bth.rgBytes[3], addr_bth.rgBytes[2], addr_bth.rgBytes[1], addr_bth.rgBytes[0] );
-        if (sockaddr_bth->port)
-            sprintf( buffer + 19, ":%lu", sockaddr_bth->port );
-        break;
-    }
+
     default:
         SetLastError( WSAEINVAL );
         return -1;
@@ -2174,26 +2192,8 @@
 int WINAPI WSAProviderConfigChange( HANDLE *handle, OVERLAPPED *overlapped,
                                     LPWSAOVERLAPPED_COMPLETION_ROUTINE completion )
 {
-    SOCKET s;
-
     FIXME( "(%p %p %p) Stub!\n", handle, overlapped, completion );
-
-    if (!handle)
-    {
-        SetLastError( WSAEFAULT );
-        return SOCKET_ERROR;
-    }
-
-    if (*handle != NULL) return 0;
-
-    s = WSASocketW( AF_INET, SOCK_STREAM, 0, NULL, 0, WSA_FLAG_OVERLAPPED );
-    if (s != INVALID_SOCKET)
-    {
-        *handle = (HANDLE)s;
-        return 0;
-    }
-
-    return SOCKET_ERROR;
+    return -1;
 }
 
 
