diff --git a/dlls/dmsynth/synthsink.c b/dlls/dmsynth/synthsink.c
--- a/dlls/dmsynth/synthsink.c
+++ b/dlls/dmsynth/synthsink.c
@@ -44,7 +44,6 @@
 
     CRITICAL_SECTION cs;
     REFERENCE_TIME latency_time;
-    REFERENCE_TIME latency;
 
     DWORD written; /* number of bytes written out */
     HANDLE stop_event;
@@ -75,23 +74,6 @@
     }
 }
 
-static HRESULT synth_sink_wait_write(struct synth_sink *sink, HANDLE buffer_event)
-{
-    HANDLE handles[] = {sink->stop_event, buffer_event};
-    DWORD ret;
-
-    ret = WaitForMultipleObjects(ARRAY_SIZE(handles), handles, FALSE, INFINITE);
-    if (ret == WAIT_FAILED)
-    {
-        ERR("WaitForSingleObject failed with %lu\n", GetLastError());
-        return HRESULT_FROM_WIN32(GetLastError());
-    }
-    if (ret == WAIT_OBJECT_0)
-        return S_FALSE;
-
-    return S_OK;
-}
-
 static HRESULT synth_sink_write_data(struct synth_sink *sink, IDirectSoundBuffer *buffer,
         DSBCAPS *caps, WAVEFORMATEX *format, const void *data, DWORD size)
 {
@@ -109,7 +91,7 @@
 
         if (FAILED(hr = IDirectSoundBuffer_GetCurrentPosition(buffer, &play_pos, &write_pos))) return hr;
 
-        if (current_pos - play_pos < write_pos - play_pos)
+        if (current_pos - play_pos <= write_pos - play_pos)
         {
             ERR("Underrun detected, sink %p, play pos %#lx, write pos %#lx, current pos %#lx!\n",
                     buffer, play_pos, write_pos, current_pos);
@@ -196,12 +178,8 @@
 static HRESULT synth_sink_render_data(struct synth_sink *sink, IDirectMusicSynth *synth,
         IDirectSoundBuffer *buffer, WAVEFORMATEX *format, short *samples, DWORD samples_size)
 {
-    REFERENCE_TIME master_time;
     REFERENCE_TIME sample_time;
     HRESULT hr;
-
-    if (FAILED(hr = IReferenceClock_GetTime(sink->master_clock, &master_time)))
-        ERR("Failed to get master clock time, hr %#lx\n", hr);
 
     if (FAILED(hr = IDirectMusicSynth_Render(synth, samples, samples_size / format->nBlockAlign,
             sink->written / format->nBlockAlign)))
@@ -212,7 +190,7 @@
         ERR("Failed to convert sample position to time, hr %#lx\n", hr);
 
     EnterCriticalSection(&sink->cs);
-    sink->latency = sample_time - master_time;
+    sink->latency_time = sample_time;
     LeaveCriticalSection(&sink->cs);
 
     return hr;
@@ -289,14 +267,17 @@
     while (SUCCEEDED(hr) && SUCCEEDED(hr = synth_sink_write_data(sink, buffer,
             &caps, &format, samples, samples_size)))
     {
-        HRESULT wait_hr;
+        HANDLE handles[] = {sink->stop_event, buffer_event};
+        DWORD ret;
 
         if (hr == S_OK) /* if successfully written, render more data */
             hr = synth_sink_render_data(sink, synth, buffer, &format, samples, samples_size);
 
-        if (S_OK != (wait_hr = synth_sink_wait_write(sink, buffer_event)))
+        if (!(ret = WaitForMultipleObjects(ARRAY_SIZE(handles), handles, FALSE, INFINITE))
+                || ret >= ARRAY_SIZE(handles))
         {
-            hr = wait_hr;
+            ERR("WaitForMultipleObjects returned %lu\n", ret);
+            hr = HRESULT_FROM_WIN32(ret);
             break;
         }
     }
@@ -715,19 +696,13 @@
 static HRESULT WINAPI latency_clock_GetTime(IReferenceClock *iface, REFERENCE_TIME *time)
 {
     struct synth_sink *This = impl_from_IReferenceClock(iface);
-    REFERENCE_TIME master_time;
-    HRESULT hr;
 
     TRACE("(%p, %p)\n", iface, time);
 
     if (!time) return E_INVALIDARG;
     if (!This->active) return E_FAIL;
 
-    if (FAILED(hr = IReferenceClock_GetTime(This->master_clock, &master_time)))
-        return hr;
-
     EnterCriticalSection(&This->cs);
-    This->latency_time = max(This->latency_time, master_time + This->latency);
     *time = This->latency_time;
     LeaveCriticalSection(&This->cs);
 
