diff --git a/dlls/oleaut32/variant.c b/dlls/oleaut32/variant.c
--- a/dlls/oleaut32/variant.c
+++ b/dlls/oleaut32/variant.c
@@ -590,7 +590,6 @@
         {
             IRecordInfo_RecordClear(rec_info, V_RECORD(pVarg));
             IRecordInfo_Release(rec_info);
-            CoTaskMemFree(V_RECORD(pVarg));
         }
         break;
     }
@@ -1187,7 +1186,7 @@
  *| 11-15  0-23    Hours (24 hour clock). 24-32 are invalid.
  */
 INT WINAPI DosDateTimeToVariantTime(USHORT wDosDate, USHORT wDosTime,
-                                    DOUBLE *pDateOut)
+                                    double *pDateOut)
 {
   UDATE ud;
 
@@ -1227,7 +1226,7 @@
  * NOTES
  *   See DosDateTimeToVariantTime() for Dos format details and bugs.
  */
-INT WINAPI VariantTimeToDosDateTime(DOUBLE dateIn, USHORT *pwDosDate, USHORT *pwDosTime)
+INT WINAPI VariantTimeToDosDateTime(double dateIn, USHORT *pwDosDate, USHORT *pwDosTime)
 {
   UDATE ud;
 
@@ -1261,7 +1260,7 @@
  *  Success: TRUE. *pDateOut contains the converted value.
  *  Failure: FALSE, if lpSt cannot be represented in VT_DATE format.
  */
-INT WINAPI SystemTimeToVariantTime(LPSYSTEMTIME lpSt, DOUBLE *pDateOut)
+INT WINAPI SystemTimeToVariantTime(LPSYSTEMTIME lpSt, double *pDateOut)
 {
   UDATE ud;
 
@@ -1292,7 +1291,7 @@
  *  Success: TRUE. *lpSt contains the converted value.
  *  Failure: FALSE, if dateIn is too large or small.
  */
-INT WINAPI VariantTimeToSystemTime(DOUBLE dateIn, LPSYSTEMTIME lpSt)
+INT WINAPI VariantTimeToSystemTime(double dateIn, LPSYSTEMTIME lpSt)
 {
   UDATE ud;
 
@@ -3107,7 +3106,7 @@
         V_VT(&varLeft) = VT_I4; /* Don't overflow */
     else
     {
-        DOUBLE d;
+        double d;
 
         if (V_VT(&varLeft) == VT_BSTR &&
             FAILED(VarR8FromStr(V_BSTR(&varLeft),
@@ -3123,7 +3122,7 @@
         V_VT(&varRight) = VT_I4; /* Don't overflow */
     else
     {
-        DOUBLE d;
+        double d;
 
         if (V_VT(&varRight) == VT_BSTR &&
             FAILED(VarR8FromStr(V_BSTR(&varRight),
@@ -4253,7 +4252,7 @@
         V_VT(&varLeft) = VT_I4; /* Don't overflow */
     else
     {
-        DOUBLE d;
+        double d;
 
         if (V_VT(&varLeft) == VT_BSTR &&
             FAILED(VarR8FromStr(V_BSTR(&varLeft), LOCALE_USER_DEFAULT, 0, &d)))
@@ -4268,7 +4267,7 @@
         V_VT(&varRight) = VT_I4; /* Don't overflow */
     else
     {
-        DOUBLE d;
+        double d;
 
         if (V_VT(&varRight) == VT_BSTR &&
             FAILED(VarR8FromStr(V_BSTR(&varRight), LOCALE_USER_DEFAULT, 0, &d)))
@@ -4615,7 +4614,7 @@
     VARTYPE vt;
     VARIANT varLeft, varRight;
     VARIANT tempLeft, tempRight;
-    DOUBLE d;
+    double d;
     HRESULT hRet;
 
     TRACE("(%s,%s,%p)\n", debugstr_variant(pVarLeft), debugstr_variant(pVarRight), pVarOut);
@@ -5199,7 +5198,7 @@
 	break;
     case VT_DECIMAL:
     {
-        DOUBLE dbl;
+        double dbl;
 
         hRet = VarR8FromDec(&V_DECIMAL(pVarIn), &dbl);
         if (FAILED(hRet))
@@ -5758,7 +5757,7 @@
     VARTYPE leftvt,rightvt;
     VARTYPE rightExtraFlags,leftExtraFlags,ExtraFlags;
     VARIANT lv,rv;
-    DOUBLE d;
+    double d;
     VARIANT tempLeft, tempRight;
 
     VariantInit(&lv);
