diff --git a/dlls/qasf/asfreader.c b/dlls/qasf/asfreader.c
--- a/dlls/qasf/asfreader.c
+++ b/dlls/qasf/asfreader.c
@@ -483,13 +483,14 @@
     struct asf_reader *filter = impl_from_strmbase_filter(iface);
     WMT_STREAM_SELECTION selections[ARRAY_SIZE(filter->streams)];
     WORD stream_numbers[ARRAY_SIZE(filter->streams)];
-    IWMReaderAdvanced *reader_advanced;
+    IWMReaderAdvanced2 *reader_advanced;
     HRESULT hr = S_OK;
+    BOOL value;
     int i;
 
     TRACE("iface %p\n", iface);
 
-    if (FAILED(hr = IWMReader_QueryInterface(filter->reader, &IID_IWMReaderAdvanced, (void **)&reader_advanced)))
+    if (FAILED(hr = IWMReader_QueryInterface(filter->reader, &IID_IWMReaderAdvanced2, (void **)&reader_advanced)))
         return hr;
 
     for (i = 0; i < filter->stream_count; ++i)
@@ -502,6 +503,14 @@
 
         if (!stream->source.pin.peer)
             continue;
+
+        value = IMemInputPin_ReceiveCanBlock(stream->source.pMemInputPin) == S_OK;
+        if (FAILED(hr = IWMReaderAdvanced2_SetOutputSetting(reader_advanced, i, L"DedicatedDeliveryThread",
+                WMT_TYPE_BOOL, (BYTE *)&value, sizeof(value))))
+        {
+            WARN("Failed to set DedicatedDeliveryThread for stream %u, hr %#lx\n", i, hr);
+            break;
+        }
 
         if (FAILED(hr = IMemAllocator_Commit(stream->source.pAllocator)))
         {
@@ -509,7 +518,7 @@
             break;
         }
 
-        if (FAILED(hr = IWMReaderAdvanced_SetAllocateForOutput(reader_advanced, i, TRUE)))
+        if (FAILED(hr = IWMReaderAdvanced2_SetAllocateForOutput(reader_advanced, i, TRUE)))
         {
             WARN("Failed to enable allocation for stream %u, hr %#lx\n", i, hr);
             break;
@@ -531,7 +540,7 @@
             break;
         }
 
-        if (FAILED(hr = IPin_NewSegment(stream->source.pin.peer, 0, 0, 1)))
+        if (FAILED(hr = IPin_NewSegment(stream->source.pin.peer, stream->seek.llCurrent, stream->seek.llStop, stream->seek.dRate)))
         {
             WARN("Failed to start stream %u new segment, hr %#lx\n", i, hr);
             break;
@@ -540,11 +549,19 @@
         selections[i] = WMT_ON;
     }
 
-    if (SUCCEEDED(hr) && FAILED(hr = IWMReaderAdvanced_SetStreamsSelected(reader_advanced,
+    if (SUCCEEDED(hr) && FAILED(hr = IWMReaderAdvanced2_SetStreamsSelected(reader_advanced,
             filter->stream_count, stream_numbers, selections)))
         WARN("Failed to set reader %p stream selection, hr %#lx\n", filter->reader, hr);
 
-    IWMReaderAdvanced_Release(reader_advanced);
+    if (SUCCEEDED(hr) && FAILED(hr = IWMReaderAdvanced2_SetUserProvidedClock(reader_advanced, !filter->filter.clock)))
+        WARN("Failed to set user provided clock, hr %#lx\n", hr);
+    else if (!filter->filter.clock)
+    {
+        if (SUCCEEDED(hr) && FAILED(hr = IWMReaderAdvanced2_DeliverTime(reader_advanced, -1)))
+            WARN("Failed to set user time, hr %#lx\n", hr);
+    }
+
+    IWMReaderAdvanced2_Release(reader_advanced);
 
     if (FAILED(hr))
         return hr;
@@ -614,7 +631,11 @@
         buffer_size = format->nAvgBytesPerSec;
     }
 
-    req_props->cBuffers = max(req_props->cBuffers, 1);
+    if (IsEqualGUID(&stream->source.pin.mt.majortype, &MEDIATYPE_Audio))
+        req_props->cBuffers = max(req_props->cBuffers, 50);
+    else
+        req_props->cBuffers = max(req_props->cBuffers, 10);
+
     req_props->cbBuffer = max(req_props->cbBuffer, buffer_size);
     req_props->cbAlign = max(req_props->cbAlign, 1);
     return IMemAllocator_SetProperties(allocator, req_props, &ret_props);
@@ -835,11 +856,20 @@
             for (i = 0; i < stream_count; ++i)
             {
                 struct asf_stream *stream = filter->streams + i;
+                const char *sgi = getenv("SteamGameId");
 
                 if (FAILED(hr = asf_stream_get_media_type(&stream->source.pin, 0, &stream_media_type)))
                     WARN("Failed to get stream media type, hr %#lx.\n", hr);
                 if (IsEqualGUID(&stream_media_type.majortype, &MEDIATYPE_Video))
-                    swprintf(name, ARRAY_SIZE(name), L"Raw Video %u", stream->index);
+                {
+                    /* King of Fighters XIII requests the WMV decoder filter pins by name
+                     * to connect them to a Sample Grabber filter.
+                     */
+                    if (sgi && (!strcmp(sgi, "222940") || !strcmp(sgi, "3050220")))
+                        swprintf(name, ARRAY_SIZE(name), L"out0");
+                    else
+                        swprintf(name, ARRAY_SIZE(name), L"Raw Video %u", stream->index);
+                }
                 else
                     swprintf(name, ARRAY_SIZE(name), L"Raw Audio %u", stream->index);
                 FreeMediaType(&stream_media_type);
