diff --git a/dlls/jscript/engine.c b/dlls/jscript/engine.c
--- a/dlls/jscript/engine.c
+++ b/dlls/jscript/engine.c
@@ -490,33 +490,48 @@
         IDispatch_Release(scope->obj);
 }
 
-static HRESULT scope_lookup_prop(jsdisp_t *jsdisp, const WCHAR *name, unsigned flags, struct property_info *desc)
-{
-    scope_chain_t *scope = scope_from_dispex(jsdisp);
-
-    return jsdisp_index_lookup(&scope->dispex, name, scope->detached_vars->argc, desc);
-}
-
-static HRESULT scope_prop_get(jsdisp_t *dispex, unsigned idx, jsval_t *r)
-{
-    scope_chain_t *scope = scope_from_dispex(dispex);
-
-    return jsval_copy(scope->detached_vars->var[idx], r);
-}
-
-static HRESULT scope_prop_put(jsdisp_t *dispex, unsigned idx, jsval_t val)
+static unsigned scope_indexed_len(jsdisp_t *dispex)
+{
+    return scope_from_dispex(dispex)->detached_vars->argc;
+}
+
+static HRESULT scope_prop_get(jsdisp_t *dispex, DISPID id, jsval_t *r)
+{
+    if(!is_indexed_prop_id(id))
+        return S_FALSE;
+    return jsval_copy(scope_from_dispex(dispex)->detached_vars->var[indexed_prop_id_to_idx(id)], r);
+}
+
+static HRESULT scope_prop_put(jsdisp_t *dispex, DISPID id, jsval_t val)
 {
     scope_chain_t *scope = scope_from_dispex(dispex);
     jsval_t copy, *ref;
     HRESULT hres;
 
+    if(!is_indexed_prop_id(id))
+        return S_FALSE;
+
     hres = jsval_copy(val, &copy);
     if(FAILED(hres))
         return hres;
 
-    ref = &scope->detached_vars->var[idx];
+    ref = &scope->detached_vars->var[indexed_prop_id_to_idx(id)];
     jsval_release(*ref);
     *ref = copy;
+    return S_OK;
+}
+
+static HRESULT scope_prop_get_desc(jsdisp_t *dispex, DISPID id, BOOL flags_only, property_desc_t *desc)
+{
+    if(!flags_only) {
+        HRESULT hres = scope_prop_get(dispex, id, &desc->value);
+        if(hres != S_OK)
+            return hres;
+    }else if(!is_indexed_prop_id(id)) {
+        return S_FALSE;
+    }
+
+    desc->flags = PROPF_ENUMERABLE | PROPF_WRITABLE;
     return S_OK;
 }
 
@@ -555,13 +570,36 @@
     return scope->obj && (jsobj = to_jsdisp(scope->obj)) ? gc_process_linked_obj(gc_ctx, op, dispex, jsobj, (void**)&scope->obj) : S_OK;
 }
 
+static void scope_cc_traverse(jsdisp_t *dispex, nsCycleCollectionTraversalCallback *cb)
+{
+    scope_chain_t *scope = scope_from_dispex(dispex);
+    note_edge_t note_edge = cc_api.note_edge;
+
+    if(scope->detached_vars) {
+        struct vars_buffer *vars = scope->detached_vars;
+        unsigned i, cnt = vars->argc;
+
+        for(i = 0; i < cnt; i++)
+            if(is_object_instance(vars->var[i]))
+                note_edge(get_edge_obj(get_object(vars->var[i])), "var", cb);
+    }
+
+    if(scope->next)
+        note_edge((IUnknown*)&scope->next->dispex.IWineJSDispatch_iface, "next", cb);
+
+    if(scope->obj)
+        note_edge(get_edge_obj(scope->obj), "obj", cb);
+}
+
 static const builtin_info_t scope_info = {
-    JSCLASS_NONE,
-    .destructor  = scope_destructor,
-    .lookup_prop = scope_lookup_prop,
-    .prop_get    = scope_prop_get,
-    .prop_put    = scope_prop_put,
-    .gc_traverse = scope_gc_traverse
+    .class         = JSCLASS_NONE,
+    .destructor    = scope_destructor,
+    .indexed_len   = scope_indexed_len,
+    .prop_get      = scope_prop_get,
+    .prop_put      = scope_prop_put,
+    .prop_get_desc = scope_prop_get_desc,
+    .gc_traverse   = scope_gc_traverse,
+    .cc_traverse   = scope_cc_traverse
 };
 
 static HRESULT scope_push(script_ctx_t *ctx, scope_chain_t *scope, IDispatch *obj, scope_chain_t **ret)
