diff --git a/dlls/mshtml/htmlframe.c b/dlls/mshtml/htmlframe.c
--- a/dlls/mshtml/htmlframe.c
+++ b/dlls/mshtml/htmlframe.c
@@ -24,7 +24,6 @@
 #include "winbase.h"
 #include "winuser.h"
 #include "ole2.h"
-#include "mshtmdid.h"
 
 #include "mshtml_private.h"
 #include "binding.h"
@@ -906,25 +905,6 @@
     .bind_to_tree          = HTMLFrameElement_bind_to_tree,
 };
 
-static void HTMLFrameElement_init_dispex_info(dispex_data_t *info, compat_mode_t mode)
-{
-    static const dispex_hook_t base2_hooks[] = {
-        {DISPID_IHTMLFRAMEBASE2_CONTENTWINDOW,      .noattr = TRUE},
-        {DISPID_IHTMLFRAMEBASE2_READYSTATE,         .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    static const dispex_hook_t hooks[] = {
-        {DISPID_IHTMLFRAMEELEMENT3_CONTENTDOCUMENT, .noattr = TRUE},
-        {DISPID_IHTMLFRAMEELEMENT3_IE8_LONGDESC,    .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    dispex_info_add_interface(info, IHTMLFrameBase_tid, NULL);
-    dispex_info_add_interface(info, IHTMLFrameBase2_tid, base2_hooks);
-    dispex_info_add_interface(info, IHTMLFrameElement3_tid, hooks);
-
-    HTMLElement_init_dispex_info(info, mode);
-}
-
 static const event_target_vtbl_t HTMLFrameElement_event_target_vtbl = {
     {
         HTMLELEMENT_DISPEX_VTBL_ENTRIES,
@@ -940,12 +920,20 @@
     .handle_event       = HTMLElement_handle_event
 };
 
+static const tid_t HTMLFrameElement_iface_tids[] = {
+    IHTMLFrameBase_tid,
+    IHTMLFrameBase2_tid,
+    IHTMLFrameElement3_tid,
+    0
+};
+
 dispex_static_data_t HTMLFrameElement_dispex = {
-    .id           = OBJID_HTMLFrameElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLFrameElement,
+    .prototype_id = PROT_HTMLElement,
     .vtbl         = &HTMLFrameElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLFrameElement_tid,
-    .init_info    = HTMLFrameElement_init_dispex_info,
+    .iface_tids   = HTMLFrameElement_iface_tids,
+    .init_info    = HTMLElement_init_dispex_info,
     .js_flags     = HOSTOBJ_VOLATILE_PROPS
 };
 
@@ -1353,25 +1341,6 @@
     .bind_to_tree          = HTMLIFrame_bind_to_tree,
 };
 
-static void HTMLIFrameElement_init_dispex_info(dispex_data_t *info, compat_mode_t mode)
-{
-    static const dispex_hook_t base2_hooks[] = {
-        {DISPID_IHTMLFRAMEBASE2_CONTENTWINDOW,       .noattr = TRUE},
-        {DISPID_IHTMLFRAMEBASE2_READYSTATE,          .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    static const dispex_hook_t hooks[] = {
-        {DISPID_IHTMLIFRAMEELEMENT3_CONTENTDOCUMENT, .noattr = TRUE},
-        {DISPID_IHTMLIFRAMEELEMENT3_IE8_LONGDESC,    .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    dispex_info_add_interface(info, IHTMLFrameBase_tid, NULL);
-    dispex_info_add_interface(info, IHTMLFrameBase2_tid, base2_hooks);
-    dispex_info_add_interface(info, IHTMLIFrameElement3_tid, hooks);
-
-    HTMLElement_init_dispex_info(info, mode);
-}
-
 static const event_target_vtbl_t HTMLIFrameElement_event_target_vtbl = {
     {
         HTMLELEMENT_DISPEX_VTBL_ENTRIES,
@@ -1388,18 +1357,21 @@
 };
 
 static const tid_t HTMLIFrameElement_iface_tids[] = {
+    IHTMLFrameBase_tid,
+    IHTMLFrameBase2_tid,
     IHTMLIFrameElement_tid,
     IHTMLIFrameElement2_tid,
+    IHTMLIFrameElement3_tid,
     0
 };
 
 dispex_static_data_t HTMLIFrameElement_dispex = {
-    .id           = OBJID_HTMLIFrameElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLIFrameElement,
+    .prototype_id = PROT_HTMLElement,
     .vtbl         = &HTMLIFrameElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLIFrame_tid,
     .iface_tids   = HTMLIFrameElement_iface_tids,
-    .init_info    = HTMLIFrameElement_init_dispex_info,
+    .init_info    = HTMLElement_init_dispex_info,
     .js_flags     = HOSTOBJ_VOLATILE_PROPS
 };
 
