diff --git a/dlls/ntoskrnl.exe/ntoskrnl.c b/dlls/ntoskrnl.exe/ntoskrnl.c
--- a/dlls/ntoskrnl.exe/ntoskrnl.c
+++ b/dlls/ntoskrnl.exe/ntoskrnl.c
@@ -30,6 +30,7 @@
 #include "evntprov.h"
 #include "ddk/csq.h"
 #include "wine/server.h"
+#include "wine/heap.h"
 #include "wine/svcctl.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(ntoskrnl);
@@ -257,6 +258,15 @@
     return header->type;
 }
 
+static const WCHAR section_type_name[] = {'S','e','c','t','i','o','n',0};
+
+static struct _OBJECT_TYPE section_type =
+{
+    section_type_name
+};
+
+static POBJECT_TYPE p_section_type = &section_type;
+
 static const POBJECT_TYPE *known_types[] =
 {
     &ExEventObjectType,
@@ -266,7 +276,8 @@
     &IoFileObjectType,
     &PsProcessType,
     &PsThreadType,
-    &SeTokenObjectType
+    &SeTokenObjectType,
+    &p_section_type,
 };
 
 DECLARE_CRITICAL_SECTION(handle_map_cs);
@@ -2121,7 +2132,7 @@
     irp->Cancel = TRUE;
     if (!(cancel_routine = IoSetCancelRoutine( irp, NULL )))
     {
-        IoReleaseCancelSpinLock( irql );
+        IoReleaseCancelSpinLock( irp->CancelIrql );
         return FALSE;
     }
 
@@ -2950,18 +2961,6 @@
 VOID WINAPI MmLockPagableSectionByHandle(PVOID ImageSectionHandle)
 {
     FIXME("stub %p\n", ImageSectionHandle);
-}
-
-/***********************************************************************
- *           MmMapLockedPages   (NTOSKRNL.EXE.@)
- */
-PVOID WINAPI MmMapLockedPages( MDL *mdl, KPROCESSOR_MODE mode )
-{
-    TRACE( "%p %u\n", mdl, mode );
-
-    mdl->MdlFlags |= MDL_MAPPED_TO_SYSTEM_VA;
-    mdl->MappedSystemVa = (char *)mdl->StartVa + mdl->ByteOffset;
-    return mdl->MappedSystemVa;
 }
 
 /***********************************************************************
@@ -4085,7 +4084,7 @@
     static const WCHAR driverW[] = {'\\','D','r','i','v','e','r','\\',0};
     WCHAR *str;
 
-    if (!(str = HeapAlloc( GetProcessHeap(), 0, sizeof(driverW) + service_name->Length - lstrlenW(servicesW)*sizeof(WCHAR) )))
+    if (!(str = heap_alloc( sizeof(driverW) + service_name->Length - lstrlenW(servicesW)*sizeof(WCHAR) )))
         return FALSE;
 
     lstrcpyW( str, driverW );
@@ -4509,18 +4508,27 @@
     reverse_barrier.TotalProcessors = cpu_count;
     cpu_count_barrier = cpu_count;
 
-    if (last_cpu_count < cpu_count)
-    {
-        struct generic_call_dpc_context *new_contexts;
-        if (!(new_contexts = realloc(contexts, sizeof(*contexts) * cpu_count)))
+    if (contexts)
+    {
+        if (last_cpu_count < cpu_count)
         {
-            ERR("No memory.\n");
-            LeaveCriticalSection(&dpc_call_cs);
-            return;
+            static struct generic_call_dpc_context *new_contexts;
+            if (!(new_contexts = heap_realloc(contexts, sizeof(*contexts) * cpu_count)))
+            {
+                ERR("No memory.\n");
+                LeaveCriticalSection(&dpc_call_cs);
+                return;
+            }
+            contexts = new_contexts;
+            SetThreadpoolThreadMinimum(dpc_call_tp, cpu_count);
+            SetThreadpoolThreadMaximum(dpc_call_tp, cpu_count);
         }
-        contexts = new_contexts;
-        SetThreadpoolThreadMinimum(dpc_call_tp, cpu_count);
-        SetThreadpoolThreadMaximum(dpc_call_tp, cpu_count);
+    }
+    else if (!(contexts = heap_alloc(sizeof(*contexts) * cpu_count)))
+    {
+        ERR("No memory.\n");
+        LeaveCriticalSection(&dpc_call_cs);
+        return;
     }
 
     memset(contexts, 0, sizeof(*contexts) * cpu_count);
