diff --git a/dlls/setupapi/install.c b/dlls/setupapi/install.c
--- a/dlls/setupapi/install.c
+++ b/dlls/setupapi/install.c
@@ -67,14 +67,6 @@
     int                 modules_size;
     int                 modules_count;
     HMODULE            *modules;
-};
-
-/* info passed to callback functions dealing with installing a section */
-struct install_section_callback_info
-{
-    HWND                hwnd;
-    PSP_FILE_CALLBACK_W callback;
-    PVOID               callback_context;
 };
 
 typedef BOOL (*iterate_fields_func)( HINF hinf, PCWSTR field, void *arg );
@@ -788,33 +780,6 @@
 }
 
 /***********************************************************************
- *            include_callback
- *
- * Called once for each Include entry in a given section.
- */
-static BOOL include_callback( HINF hinf, PCWSTR field, void *arg )
-{
-    SetupOpenAppendInfFileW( field, hinf, NULL );
-    return TRUE;
-}
-
-/***********************************************************************
- *            install_section_callback
- *
- * Called once for each Needs entry in a given section.
- */
-static BOOL install_section_callback( HINF hinf, PCWSTR field, void *arg )
-{
-    struct install_section_callback_info *info = arg;
-
-    TRACE( "installing %s\n", debugstr_w(field) );
-    SetupInstallFromInfSectionW( info->hwnd, hinf, field, SPINST_ALL, NULL, NULL, SP_COPY_NEWER,
-                                 info->callback, info->callback_context, NULL, NULL );
-    return TRUE;
-}
-
-
-/***********************************************************************
  *            update_ini_callback
  *
  * Called once for each UpdateInis entry in a given section.
@@ -1265,8 +1230,20 @@
  */
 void WINAPI InstallHinfSectionW( HWND hwnd, HINSTANCE handle, LPCWSTR cmdline, INT show )
 {
-    struct install_section_callback_info info = { .hwnd = hwnd, .callback = SetupDefaultQueueCallbackW };
-    WCHAR *s, *path, section[MAX_PATH + ARRAY_SIZE( L".Services" )];
+#ifdef __i386__
+    static const WCHAR nt_platformW[] = L".ntx86";
+#elif defined(__x86_64__)
+    static const WCHAR nt_platformW[] = L".ntamd64";
+#elif defined(__arm__)
+    static const WCHAR nt_platformW[] = L".ntarm";
+#elif defined(__aarch64__)
+    static const WCHAR nt_platformW[] = L".ntarm64";
+#else  /* FIXME: other platforms */
+    static const WCHAR nt_platformW[] = L".nt";
+#endif
+
+    WCHAR *s, *path, section[MAX_PATH + ARRAY_SIZE( nt_platformW ) + ARRAY_SIZE( L".Services" )];
+    void *callback_context;
     UINT mode;
     HINF hinf;
 
@@ -1287,15 +1264,27 @@
     hinf = SetupOpenInfFileW( path, NULL, INF_STYLE_WIN4, NULL );
     if (hinf == INVALID_HANDLE_VALUE) return;
 
-    SetupDiGetActualSectionToInstallW( hinf, section, section, ARRAY_SIZE(section), NULL, NULL );
-    TRACE( "using section %s\n", debugstr_w(section) );
-
-    iterate_section_fields( hinf, section, L"Include", include_callback, NULL );
-
-    info.callback_context = SetupInitDefaultQueueCallback( hwnd );
-    install_section_callback( hinf, section, &info );
-    iterate_section_fields( hinf, section, L"Needs", install_section_callback, &info );
-    SetupTermDefaultQueueCallback( info.callback_context );
+    if (!(GetVersion() & 0x80000000))
+    {
+        INFCONTEXT context;
+
+        /* check for <section>.ntx86 (or corresponding name for the current platform)
+         * and then <section>.nt */
+        s = section + lstrlenW(section);
+        lstrcpyW( s, nt_platformW );
+        if (!(SetupFindFirstLineW( hinf, section, NULL, &context )))
+        {
+            lstrcpyW( s, L".nt" );
+            if (!(SetupFindFirstLineW( hinf, section, NULL, &context ))) *s = 0;
+        }
+        if (*s) TRACE( "using section %s instead\n", debugstr_w(section) );
+    }
+
+    callback_context = SetupInitDefaultQueueCallback( hwnd );
+    SetupInstallFromInfSectionW( hwnd, hinf, section, SPINST_ALL, NULL, NULL, SP_COPY_NEWER,
+                                 SetupDefaultQueueCallbackW, callback_context,
+                                 NULL, NULL );
+    SetupTermDefaultQueueCallback( callback_context );
     lstrcatW( section, L".Services" );
     SetupInstallServicesFromInfSectionW( hinf, section, 0 );
     SetupCloseInfFile( hinf );
