diff --git a/dlls/dinput/tests/hid.c b/dlls/dinput/tests/hid.c
--- a/dlls/dinput/tests/hid.c
+++ b/dlls/dinput/tests/hid.c
@@ -55,7 +55,6 @@
 #include "dinput.h"
 #include "dinputd.h"
 #include "hidusage.h"
-#include "gameinput.h"
 
 #include "wine/mssign.h"
 
@@ -720,7 +719,8 @@
                            NULL, OPEN_EXISTING, 0, NULL );
     ok( control != INVALID_HANDLE_VALUE, "CreateFile failed, error %lu\n", GetLastError() );
     ret = sync_ioctl( control, IOCTL_WINETEST_REMOVE_DEVICE, desc, sizeof(*desc), NULL, 0, 5000 );
-    ok( ret || GetLastError() == ERROR_FILE_NOT_FOUND, "IOCTL_WINETEST_REMOVE_DEVICE failed, last error %lu\n", GetLastError() );
+    ok( ret || (GetLastError() == ERROR_FILE_NOT_FOUND || GetLastError() == ERROR_NO_SUCH_DEVICE),
+        "IOCTL_WINETEST_REMOVE_DEVICE failed, last error %lu\n", GetLastError() );
     CloseHandle( control );
 
     if (!ret) return;
@@ -1618,24 +1618,6 @@
         { .DataIndex = 32, .RawValue = 0xfeedcafe, },
         { .DataIndex = 37, .RawValue = 1, },
         { .DataIndex = 39, .RawValue = 1, },
-    };
-    const BYTE expect_report_0[] =
-    {
-        report_id, 0x02, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-        0x00,      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-        0x00,      0x00, 0x00, 0x00, 0x00, 0x00
-    };
-    const BYTE expect_report_1[] =
-    {
-        report_id, 0x02, 0x03, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00,
-        0x00,      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-        0x00,      0x00, 0x00, 0x00, 0x00, 0x00
-    };
-    const BYTE expect_report_2[] =
-    {
-        report_id, 0x02, 0x03, 0x01, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x02, 0x01, 0x00,
-        0x00,      0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
-        0x00,      0x00, 0x00, 0x00, 0x00, 0x00
     };
 
     DWORD waveform_list, collection_count, generic_output_list;
@@ -2123,67 +2105,6 @@
     status = HidP_GetData( HidP_Input, data, &value, preparsed_data, report, caps.InputReportByteLength );
     ok( status == HIDP_STATUS_BUFFER_TOO_SMALL, "HidP_GetData returned %#lx\n", status );
     ok( value == 11, "got data count %ld, expected %d\n", value, 11 );
-    memset( data, 0, sizeof(data) );
-    status = HidP_GetData( HidP_Input, data, &value, preparsed_data, report, caps.InputReportByteLength );
-    ok( status == HIDP_STATUS_SUCCESS, "HidP_GetData returned %#lx\n", status );
-    for (i = 0; i < ARRAY_SIZE(expect_data); ++i)
-    {
-        winetest_push_context( "data[%d]", i );
-        check_member( data[i], expect_data[i], "%d", DataIndex );
-        check_member( data[i], expect_data[i], "%ld", RawValue );
-        winetest_pop_context();
-    }
-
-    memset( report, 0, caps.InputReportByteLength );
-    report[0] = report_id;
-
-    value = 0;
-    status = HidP_SetData( HidP_Input, data, &value, preparsed_data, report, caps.InputReportByteLength );
-    ok( status == HIDP_STATUS_REPORT_DOES_NOT_EXIST, "HidP_SetData returned %#lx\n", status );
-    ok( value == 0, "got data count %ld, expected %d\n", value, 0 );
-
-    value = 0;
-    data[value].DataIndex = 0;  data[value++].RawValue = 1;
-    data[value].DataIndex = 1;  data[value++].RawValue = 2;
-    data[value].DataIndex = 0;  data[value++].RawValue = 3;
-    data[value].DataIndex = 43; data[value++].RawValue = 0;
-    data[value].DataIndex = 2;  data[value++].RawValue = 5;
-    status = HidP_SetData( HidP_Input, data, &value, preparsed_data, report, caps.InputReportByteLength );
-    ok( status == HIDP_STATUS_IS_VALUE_ARRAY, "HidP_SetData returned %#lx\n", status );
-    ok( value == 3, "got data count %ld, expected %d\n", value, 3 );
-    ok( !memcmp( report, expect_report_0, caps.InputReportByteLength ), "report mismatch\n" );
-
-    value = 0;
-    data[value].DataIndex = 18; data[value++].RawValue = 1;
-    data[value].DataIndex = 19; data[value++].RawValue = 1;
-    data[value].DataIndex = 18; data[value++].RawValue = 0;
-    data[value].DataIndex = 18; data[value++].RawValue = 0;
-    data[value].DataIndex = 20; data[value++].RawValue = 1;
-    status = HidP_SetData( HidP_Input, data, &value, preparsed_data, report, caps.InputReportByteLength );
-    ok( status == HIDP_STATUS_BUTTON_NOT_PRESSED, "HidP_SetData returned %#lx\n", status );
-    ok( value == 3, "got data count %ld, expected %d\n", value, 3 );
-    ok( !memcmp( report, expect_report_1, caps.InputReportByteLength ), "report mismatch\n" );
-
-    value = 0;
-    data[value].DataIndex = 2;  data[value++].RawValue = 1;
-    data[value].DataIndex = 2;  data[value++].RawValue = 1;
-    data[value].DataIndex = 18; data[value++].RawValue = 1;
-    data[value].DataIndex = 18; data[value++].RawValue = 1;
-    data[value].DataIndex = 20; data[value++].RawValue = 1;
-    status = HidP_SetData( HidP_Input, data, &value, preparsed_data, report, caps.InputReportByteLength );
-    ok( status == HIDP_STATUS_BUFFER_TOO_SMALL, "HidP_SetData returned %#lx\n", status );
-    ok( value == 3, "got data count %ld, expected %d\n", value, 3 );
-    ok( !memcmp( report, expect_report_2, caps.InputReportByteLength ), "report mismatch\n" );
-
-    memset( report, 0, caps.InputReportByteLength );
-    report[0] = report_id;
-    value = ARRAY_SIZE(expect_data);
-    memcpy( data, expect_data, sizeof(expect_data) );
-    status = HidP_SetData( HidP_Input, data, &value, preparsed_data, report, caps.InputReportByteLength );
-    ok( status == HIDP_STATUS_SUCCESS, "HidP_SetData returned %#lx\n", status );
-    ok( value == ARRAY_SIZE(expect_data), "got data count %ld, expected %d\n", value, 5 );
-
-    value = ARRAY_SIZE(expect_data);
     memset( data, 0, sizeof(data) );
     status = HidP_GetData( HidP_Input, data, &value, preparsed_data, report, caps.InputReportByteLength );
     ok( status == HIDP_STATUS_SUCCESS, "HidP_GetData returned %#lx\n", status );
