diff --git a/dlls/mshtml/mutation.c b/dlls/mshtml/mutation.c
--- a/dlls/mshtml/mutation.c
+++ b/dlls/mshtml/mutation.c
@@ -438,6 +438,7 @@
 static void set_document_mode(HTMLDocumentNode *doc, compat_mode_t document_mode, BOOL emulate_mode, BOOL lock)
 {
     compat_mode_t max_compat_mode;
+    const char *sgi;
 
     if(doc->document_mode_locked) {
         WARN("attempting to set document mode %d on locked document %p\n", document_mode, doc);
@@ -454,6 +455,9 @@
              document_mode, max_compat_mode);
         document_mode = max_compat_mode;
     }
+
+    if ((sgi = getenv("SteamGameId")) && (!strcmp(sgi, "39210")))
+        document_mode = COMPAT_MODE_IE11;
 
     doc->document_mode = document_mode;
     doc->emulate_mode = emulate_mode;
@@ -1198,7 +1202,7 @@
     free(This);
 }
 
-static HRESULT init_mutation_observer_ctor(struct constructor*);
+static HRESULT create_mutation_observer_ctor(HTMLInnerWindow *script_global, DispatchEx **ret);
 
 static const dispex_static_data_vtbl_t mutation_observer_dispex_vtbl = {
     .query_interface  = mutation_observer_query_interface,
@@ -1212,12 +1216,12 @@
     0
 };
 dispex_static_data_t MutationObserver_dispex = {
-    .id               = OBJID_MutationObserver,
-    .init_constructor = init_mutation_observer_ctor,
+    .id               = PROT_MutationObserver,
+    .init_constructor = create_mutation_observer_ctor,
     .vtbl             = &mutation_observer_dispex_vtbl,
     .disp_tid         = IWineMSHTMLMutationObserver_tid,
     .iface_tids       = mutation_observer_iface_tids,
-    .min_compat_mode  = COMPAT_MODE_IE11,
+    .min_compat_mode  = COMPAT_MODE_IE11 + 1,  /* FIXME HACK: Not exposed as FFXIV Launcher breaks with MutationObserver stub */
 };
 
 static HRESULT create_mutation_observer(DispatchEx *owner, IDispatch *callback,
@@ -1243,11 +1247,26 @@
     return S_OK;
 }
 
+struct mutation_observer_ctor {
+    DispatchEx dispex;
+};
+
+static inline struct mutation_observer_ctor *mutation_observer_ctor_from_DispatchEx(DispatchEx *iface)
+{
+    return CONTAINING_RECORD(iface, struct mutation_observer_ctor, dispex);
+}
+
+static void mutation_observer_ctor_destructor(DispatchEx *dispex)
+{
+    struct mutation_observer_ctor *This = mutation_observer_ctor_from_DispatchEx(dispex);
+    free(This);
+}
+
 static HRESULT mutation_observer_ctor_value(DispatchEx *dispex, LCID lcid,
         WORD flags, DISPPARAMS *params, VARIANT *res, EXCEPINFO *ei,
         IServiceProvider *caller)
 {
-    struct constructor *This = constructor_from_DispatchEx(dispex);
+    struct mutation_observer_ctor *This = mutation_observer_ctor_from_DispatchEx(dispex);
     VARIANT *callback;
     IWineMSHTMLMutationObserver *mutation_observer;
     HRESULT hres;
@@ -1290,21 +1309,30 @@
 }
 
 static const dispex_static_data_vtbl_t mutation_observer_ctor_dispex_vtbl = {
-    .destructor       = constructor_destructor,
-    .traverse         = constructor_traverse,
-    .unlink           = constructor_unlink,
+    .destructor       = mutation_observer_ctor_destructor,
     .value            = mutation_observer_ctor_value
 };
 
 static dispex_static_data_t mutation_observer_ctor_dispex = {
-    .name           = "MutationObserver",
-    .constructor_id = OBJID_MutationObserver,
+    .name           = "Function",
+    .constructor_id = PROT_MutationObserver,
     .vtbl           = &mutation_observer_ctor_dispex_vtbl,
 };
 
-static HRESULT init_mutation_observer_ctor(struct constructor *constr)
-{
-    init_dispatch(&constr->dispex, &mutation_observer_ctor_dispex, constr->window,
-                  dispex_compat_mode(&constr->window->event_target.dispex));
+static HRESULT create_mutation_observer_ctor(HTMLInnerWindow *script_global, DispatchEx **ret)
+{
+    struct mutation_observer_ctor *obj;
+
+    obj = calloc(1, sizeof(*obj));
+    if(!obj)
+    {
+        ERR("No memory.\n");
+        return E_OUTOFMEMORY;
+    }
+
+    init_dispatch(&obj->dispex, &mutation_observer_ctor_dispex, script_global,
+                  dispex_compat_mode(&script_global->event_target.dispex));
+
+    *ret = &obj->dispex;
     return S_OK;
 }
