diff --git a/dlls/win32u/hook.c b/dlls/win32u/hook.c
--- a/dlls/win32u/hook.c
+++ b/dlls/win32u/hook.c
@@ -31,7 +31,9 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(hook);
 
-static const char * const hook_names[NB_HOOKS] =
+#define WH_WINEVENT (WH_MAXHOOK+1)
+
+static const char * const hook_names[WH_WINEVENT - WH_MINHOOK + 1] =
 {
     "WH_MSGFILTER",
     "WH_JOURNALRECORD",
@@ -137,7 +139,7 @@
     }
     SERVER_END_REQ;
 
-    TRACE( "%s %p %x -> %p\n", debugstr_hook_id(id), proc, tid, handle );
+    TRACE( "%s %p %x -> %p\n", debugstr_hook_id(id), proc, (int)tid, handle );
     return handle;
 }
 
@@ -159,10 +161,8 @@
     return !status;
 }
 
-/***********************************************************************
- *	     NtUserUnhookWindowsHook   (win32u.@)
- */
-BOOL WINAPI NtUserUnhookWindowsHook( INT id, HOOKPROC proc )
+/* see UnhookWindowsHook */
+BOOL unhook_windows_hook( INT id, HOOKPROC proc )
 {
     NTSTATUS status;
 
@@ -215,7 +215,7 @@
         h_extra.lparam = info->lparam;
 
         TRACE( "calling hook in thread %04x %s code %x wp %lx lp %lx\n",
-               info->tid, hook_names[info->id-WH_MINHOOK],
+               (int)info->tid, hook_names[info->id-WH_MINHOOK],
                info->code, (long)info->wparam, info->lparam );
 
         switch(info->id)
@@ -244,9 +244,8 @@
         HHOOK prev = thread_info->hook;
         BOOL prev_unicode = thread_info->hook_unicode;
         struct win_hook_params *params = info;
-        void *ret_ptr, *extra_buffer = NULL;
-        SIZE_T extra_buffer_size = 0;
         size_t reply_size;
+        void *ret_ptr;
         ULONG ret_len;
 
         size = FIELD_OFFSET( struct win_hook_params, module[module ? lstrlenW( module ) + 1 : 1] );
@@ -296,21 +295,21 @@
                     CBT_CREATEWNDW *cbtc = (CBT_CREATEWNDW *)params->lparam;
                     LPARAM lp = (LPARAM)cbtc->lpcs;
                     pack_user_message( (char *)params + message_offset, message_size,
-                                       WM_CREATE, 0, lp, FALSE, &extra_buffer );
+                                       WM_CREATE, 0, lp, FALSE );
                 }
                 break;
             case WH_CALLWNDPROC:
                 {
                     CWPSTRUCT *cwp = (CWPSTRUCT *)((char *)params + lparam_offset);
                     pack_user_message( (char *)params + message_offset, message_size,
-                                       cwp->message, cwp->wParam, cwp->lParam, ansi, &extra_buffer );
+                                       cwp->message, cwp->wParam, cwp->lParam, ansi );
                 }
                 break;
             case WH_CALLWNDPROCRET:
                 {
                     CWPRETSTRUCT *cwpret = (CWPRETSTRUCT *)((char *)params + lparam_offset);
                     pack_user_message( (char *)params + message_offset, message_size,
-                                       cwpret->message, cwpret->wParam, cwpret->lParam, ansi, &extra_buffer );
+                                       cwpret->message, cwpret->wParam, cwpret->lParam, ansi );
                 }
                 break;
             }
@@ -324,8 +323,6 @@
         {
             WARN("Too many hooks called recursively, skipping call.\n");
             if (params != info) free( params );
-            if (extra_buffer)
-                NtFreeVirtualMemory( GetCurrentProcess(), &extra_buffer, &extra_buffer_size, MEM_RELEASE );
             return 0;
         }
 
@@ -349,8 +346,6 @@
         thread_info->hook_call_depth--;
 
         if (params != info) free( params );
-        if (extra_buffer)
-            NtFreeVirtualMemory( GetCurrentProcess(), &extra_buffer, &extra_buffer_size, MEM_RELEASE );
     }
 
     return ret;
@@ -560,7 +555,7 @@
     ULONG ret_len;
     BOOL ret;
 
-    TRACE( "%04x, %p, %d, %d\n", event, hwnd, object_id, child_id );
+    TRACE( "%04x, %p, %d, %d\n", (int)event, hwnd, (int)object_id, (int)child_id );
 
     user_check_not_lock();
 
@@ -604,7 +599,7 @@
     do
     {
         TRACE( "calling WH_WINEVENT hook %p event %x hwnd %p %x %x module %s\n",
-               info.proc, event, hwnd, object_id, child_id, debugstr_w(info.module) );
+               info.proc, (int)event, hwnd, (int)object_id, (int)child_id, debugstr_w(info.module) );
 
         info.time = NtGetTickCount();
         KeUserModeCallback( NtUserCallWinEventHook, &info,
