diff --git a/dlls/dbghelp/module.c b/dlls/dbghelp/module.c
--- a/dlls/dbghelp/module.c
+++ b/dlls/dbghelp/module.c
@@ -31,6 +31,7 @@
 #include "psapi.h"
 #include "winternl.h"
 #include "wine/debug.h"
+#include "wine/heap.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(dbghelp);
 
@@ -83,8 +84,33 @@
 static BOOL is_wine_loader(const WCHAR *module)
 {
     const WCHAR *filename = get_filename(module, NULL);
-
-    return !wcscmp( filename, L"wine" );
+    const char *ptr;
+    BOOL ret = FALSE;
+    WCHAR *buffer;
+    DWORD len;
+
+    if ((ptr = getenv("WINELOADER")))
+    {
+        ptr = file_nameA(ptr);
+        len = 2 + MultiByteToWideChar( CP_UNIXCP, 0, ptr, -1, NULL, 0 );
+        buffer = heap_alloc( len * sizeof(WCHAR) );
+        MultiByteToWideChar( CP_UNIXCP, 0, ptr, -1, buffer, len );
+    }
+    else
+    {
+        buffer = heap_alloc( sizeof(L"wine") + 2 * sizeof(WCHAR) );
+        lstrcpyW( buffer, L"wine" );
+    }
+
+    if (!wcscmp( filename, buffer ))
+        ret = TRUE;
+
+    lstrcatW( buffer, L"64" );
+    if (!wcscmp( filename, buffer ))
+        ret = TRUE;
+
+    heap_free( buffer );
+    return ret;
 }
 
 static void module_fill_module(const WCHAR* in, WCHAR* out, size_t size)
@@ -109,9 +135,40 @@
     module_fill_module(name, module->modulename, ARRAY_SIZE(module->modulename));
 }
 
-const WCHAR *get_wine_loader_name(struct process *pcs)
-{
-    return process_getenv(pcs, L"WINELOADER");
+/* Returned string must be freed by caller */
+WCHAR *get_wine_loader_name(struct process *pcs)
+{
+    const WCHAR *name;
+    WCHAR* altname;
+    unsigned len;
+
+    name = process_getenv(pcs, L"WINELOADER");
+    if (!name) name = pcs->is_host_64bit ? L"wine64" : L"wine";
+    len = lstrlenW(name);
+
+    /* WINELOADER isn't properly updated in Wow64 process calling inside Windows env block
+     * (it's updated in ELF env block though)
+     * So do the adaptation ourselves.
+     */
+    altname = HeapAlloc(GetProcessHeap(), 0, (len + 2 + 1) * sizeof(WCHAR));
+    if (altname)
+    {
+        memcpy(altname, name, len * sizeof(WCHAR));
+        if (pcs->is_host_64bit && len >= 2 && memcmp(name + len - 2, L"64", 2 * sizeof(WCHAR)) != 0)
+        {
+            lstrcpyW(altname + len, L"64");
+            /* in multi-arch wow configuration, wine64 doesn't exist */
+            if (GetFileAttributesW(altname) == INVALID_FILE_ATTRIBUTES)
+                altname[len] = L'\0';
+        }
+        else if (!pcs->is_host_64bit && len >= 2 && !memcmp(name + len - 2, L"64", 2 * sizeof(WCHAR)))
+            altname[len - 2] = '\0';
+        else
+            altname[len] = '\0';
+    }
+
+    TRACE("returning %s\n", debugstr_w(altname));
+    return altname;
 }
 
 static const char*      get_module_type(struct module* module)
@@ -338,7 +395,7 @@
             idslW64.hFile = INVALID_HANDLE_VALUE;
 
             pcs_callback(module->process, CBA_DEFERRED_SYMBOL_LOAD_START, &idslW64);
-            ret = pe_load_debug_info(module);
+            ret = pe_load_debug_info(module->process, module);
             pcs_callback(module->process,
                          ret ? CBA_DEFERRED_SYMBOL_LOAD_COMPLETE : CBA_DEFERRED_SYMBOL_LOAD_FAILURE,
                          &idslW64);
@@ -445,7 +502,7 @@
 
     path = get_dos_file_name(file);
     handle = CreateFileW(path, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, NULL);
-    HeapFree(GetProcessHeap(), 0, path);
+    heap_free(path);
     if (handle == INVALID_HANDLE_VALUE) return FALSE;
 
     crc = calc_crc32(handle);
@@ -477,7 +534,7 @@
 
     path = get_dos_file_name(file);
     handle = CreateFileW(path, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, NULL);
-    HeapFree(GetProcessHeap(), 0, path);
+    heap_free(path);
     if (handle == INVALID_HANDLE_VALUE) return FALSE;
 
     TRACE("Located debug information file at %s\n", debugstr_w(file));
