diff --git a/dlls/quartz/vmr7.c b/dlls/quartz/vmr7.c
--- a/dlls/quartz/vmr7.c
+++ b/dlls/quartz/vmr7.c
@@ -71,8 +71,6 @@
     IVMRSurfaceAllocator *allocator;
     IVMRImagePresenter *presenter;
     IDirectDrawSurface7 **surfaces;
-    IDirectDraw7 *ddraw;
-    HMONITOR monitor;
     DWORD surface_count;
     DWORD surface_index;
     DWORD_PTR cookie;
@@ -257,7 +255,7 @@
         copy_plane(&dst, surface_desc.lPitch / 2, surface_desc.dwHeight / 2, &src, src_pitch / 2, height / 2);
         copy_plane(&dst, surface_desc.lPitch / 2, surface_desc.dwHeight / 2, &src, src_pitch / 2, height / 2);
     }
-    else if (height > 0 && (bitmap_header->biCompression == BI_RGB || bitmap_header->biCompression == BI_BITFIELDS))
+    else if (height > 0 && bitmap_header->biCompression == BI_RGB)
     {
         BYTE *dst = surface_desc.lpSurface;
         const BYTE *src = data;
@@ -285,44 +283,8 @@
     return IVMRImagePresenter_PresentImage(filter->presenter, filter->cookie, &info);
 }
 
-static BOOL fourcc_is_supported(IDirectDraw7 *ddraw, DWORD fourcc)
-{
-    DWORD *codes, count, i;
-    HRESULT hr;
-
-    if (FAILED(hr = IDirectDraw7_GetFourCCCodes(ddraw, &count, NULL)))
-    {
-        ERR("Failed to get FOURCC code count, hr %#lx.\n", hr);
-        return FALSE;
-    }
-
-    if (!count || !(codes = calloc(count, sizeof(*codes))))
-        return FALSE;
-
-    if (FAILED(hr = IDirectDraw7_GetFourCCCodes(ddraw, &count, codes)))
-    {
-        ERR("Failed to get FOURCC codes, hr %#lx.\n", hr);
-        free(codes);
-        return FALSE;
-    }
-
-    for (i = 0; i < count; ++i)
-    {
-        if (codes[i] == fourcc)
-            break;
-    }
-    free(codes);
-
-    return i < count;
-}
-
 static HRESULT vmr_query_accept(struct strmbase_renderer *iface, const AM_MEDIA_TYPE *mt)
 {
-    struct vmr7 *filter = impl_from_IBaseFilter(&iface->filter.IBaseFilter_iface);
-    const BITMAPINFOHEADER *bitmap_header = get_bitmap_header(mt);
-    IDirectDraw7 *ddraw = filter->ddraw;
-    HRESULT hr = S_OK;
-
     if (!IsEqualIID(&mt->majortype, &MEDIATYPE_Video) || !mt->pbFormat)
         return S_FALSE;
 
@@ -330,22 +292,7 @@
             && !IsEqualGUID(&mt->formattype, &FORMAT_VideoInfo2))
         return S_FALSE;
 
-    if (bitmap_header->biCompression == BI_RGB || bitmap_header->biCompression == BI_BITFIELDS)
-        return S_OK;
-
-    if (!ddraw)
-    {
-        if (FAILED(DirectDrawCreateEx(NULL, (void **)&ddraw, &IID_IDirectDraw7, NULL)))
-            return S_FALSE;
-    }
-
-    if (!fourcc_is_supported(ddraw, bitmap_header->biCompression))
-        hr = S_FALSE;
-
-    if (ddraw != filter->ddraw)
-        IDirectDraw7_Release(ddraw);
-
-    return hr;
+    return S_OK;
 }
 
 static HRESULT initialize_device(struct vmr7 *filter, VMRALLOCATIONINFO *info, DWORD count)
@@ -826,7 +773,6 @@
             }
             IUnknown_QueryInterface(default_presenter, &IID_IVMRSurfaceAllocator, (void **)&filter->allocator);
             IUnknown_QueryInterface(default_presenter, &IID_IVMRImagePresenter, (void **)&filter->presenter);
-            IVMRSurfaceAllocator_AdviseNotify(filter->allocator, &filter->IVMRSurfaceAllocatorNotify_iface);
             IUnknown_Release(default_presenter);
             break;
 
@@ -1314,19 +1260,8 @@
 static HRESULT WINAPI surface_allocator_notify_SetDDrawDevice(
         IVMRSurfaceAllocatorNotify *iface, IDirectDraw7 *device, HMONITOR monitor)
 {
-    struct vmr7 *filter = impl_from_IVMRSurfaceAllocatorNotify(iface);
-
-    TRACE("filter %p, device %p, monitor %p.\n", filter, device, monitor);
-
-    if (!device || monitor == MONITOR_DEFAULTTONULL)
-    {
-        WARN("Invalid parameters.\n");
-        return E_FAIL;
-    }
-
-    filter->ddraw = device;
-    filter->monitor = monitor;
-    return S_OK;
+    FIXME("iface %p, device %p, monitor %p, stub!\n", iface, device, monitor);
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI surface_allocator_notify_ChangeDDrawDevice(
