diff --git a/dlls/dsound/tests/dsound8.c b/dlls/dsound/tests/dsound8.c
--- a/dlls/dsound/tests/dsound8.c
+++ b/dlls/dsound/tests/dsound8.c
@@ -728,7 +728,7 @@
 
         for (f = 0; f < ARRAY_SIZE(formats); f++) {
           for (tag = 0; tag < ARRAY_SIZE(format_tags); tag++) {
-            WAVEFORMATEXTENSIBLE *wfxe;
+            WAVEFORMATEXTENSIBLE wfxe;
 
             /* if float, we only want to test 32-bit */
             if ((format_tags[tag] == WAVE_FORMAT_IEEE_FLOAT) && (formats[f][1] != 32))
@@ -770,14 +770,13 @@
                 IDirectSoundBuffer_Release(secondary);
             secondary = NULL;
 
-            wfxe = calloc(1, sizeof(wfx) + sizeof(*wfxe) + 1);
-            bufdesc.lpwfxFormat = (WAVEFORMATEX*)wfxe;
-            wfxe->Format = wfx;
-            wfxe->Format.wFormatTag = WAVE_FORMAT_EXTENSIBLE;
-            wfxe->SubFormat = (format_tags[tag] == WAVE_FORMAT_PCM ? KSDATAFORMAT_SUBTYPE_PCM : KSDATAFORMAT_SUBTYPE_IEEE_FLOAT);
-            wfxe->Format.cbSize = 1;
-            wfxe->Samples.wValidBitsPerSample = wfx.wBitsPerSample;
-            wfxe->dwChannelMask = (wfx.nChannels == 1 ? KSAUDIO_SPEAKER_MONO : KSAUDIO_SPEAKER_STEREO);
+            bufdesc.lpwfxFormat=(WAVEFORMATEX*)&wfxe;
+            wfxe.Format = wfx;
+            wfxe.Format.wFormatTag = WAVE_FORMAT_EXTENSIBLE;
+            wfxe.SubFormat = (format_tags[tag] == WAVE_FORMAT_PCM ? KSDATAFORMAT_SUBTYPE_PCM : KSDATAFORMAT_SUBTYPE_IEEE_FLOAT);
+            wfxe.Format.cbSize = 1;
+            wfxe.Samples.wValidBitsPerSample = wfx.wBitsPerSample;
+            wfxe.dwChannelMask = (wfx.nChannels == 1 ? KSAUDIO_SPEAKER_MONO : KSAUDIO_SPEAKER_STEREO);
 
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok(rc==DSERR_INVALIDPARAM && !secondary,
@@ -789,7 +788,7 @@
                 secondary=NULL;
             }
 
-            wfxe->Format.cbSize = sizeof(*wfxe) - sizeof(wfx) + 1;
+            wfxe.Format.cbSize = sizeof(wfxe) - sizeof(wfx) + 1;
 
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok(((rc==DSERR_CONTROLUNAVAIL || rc==DSERR_INVALIDCALL /* 2003 */ || rc==DSERR_INVALIDPARAM) && !secondary)
@@ -802,8 +801,8 @@
                 secondary=NULL;
             }
 
-            wfxe->Format.cbSize = sizeof(*wfxe) - sizeof(wfx);
-            wfxe->SubFormat = GUID_NULL;
+            wfxe.Format.cbSize = sizeof(wfxe) - sizeof(wfx);
+            wfxe.SubFormat = GUID_NULL;
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok((rc==DSERR_INVALIDPARAM || rc==DSERR_INVALIDCALL) && !secondary,
                 "IDirectSound_CreateSoundBuffer() returned: %08lx %p\n",
@@ -814,7 +813,7 @@
                 secondary=NULL;
             }
 
-            wfxe->Format.cbSize = sizeof(*wfxe);
+            wfxe.Format.cbSize = sizeof(wfxe);
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok((rc==DSERR_CONTROLUNAVAIL || rc==DSERR_INVALIDCALL || rc==DSERR_INVALIDPARAM) && !secondary,
                 "IDirectSound_CreateSoundBuffer() returned: %08lx %p\n",
@@ -825,7 +824,7 @@
                 secondary=NULL;
             }
 
-            wfxe->SubFormat = (format_tags[tag] == WAVE_FORMAT_PCM ? KSDATAFORMAT_SUBTYPE_PCM : KSDATAFORMAT_SUBTYPE_IEEE_FLOAT);
+            wfxe.SubFormat = (format_tags[tag] == WAVE_FORMAT_PCM ? KSDATAFORMAT_SUBTYPE_PCM : KSDATAFORMAT_SUBTYPE_IEEE_FLOAT);
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok(rc==DS_OK && secondary,
                 "IDirectSound_CreateSoundBuffer() returned: %08lx %p\n",
@@ -836,7 +835,7 @@
                 secondary=NULL;
             }
 
-            wfxe->Format.cbSize = sizeof(*wfxe) + 1;
+            wfxe.Format.cbSize = sizeof(wfxe) + 1;
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok(((rc==DSERR_CONTROLUNAVAIL || rc==DSERR_INVALIDCALL /* 2003 */ || rc==DSERR_INVALIDPARAM) && !secondary)
                 || rc==DS_OK /* driver dependent? */,
@@ -848,8 +847,8 @@
                 secondary=NULL;
             }
 
-            wfxe->Format.cbSize = sizeof(*wfxe) - sizeof(wfx);
-            ++wfxe->Samples.wValidBitsPerSample;
+            wfxe.Format.cbSize = sizeof(wfxe) - sizeof(wfx);
+            ++wfxe.Samples.wValidBitsPerSample;
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok(rc==DSERR_INVALIDPARAM && !secondary,
                 "IDirectSound_CreateSoundBuffer() returned: %08lx %p\n",
@@ -859,9 +858,9 @@
                 IDirectSoundBuffer_Release(secondary);
                 secondary=NULL;
             }
-            --wfxe->Samples.wValidBitsPerSample;
-
-            wfxe->Samples.wValidBitsPerSample = 0;
+            --wfxe.Samples.wValidBitsPerSample;
+
+            wfxe.Samples.wValidBitsPerSample = 0;
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok(rc==DS_OK && secondary,
                 "IDirectSound_CreateSoundBuffer() returned: %08lx %p\n",
@@ -871,7 +870,7 @@
                 IDirectSoundBuffer_Release(secondary);
                 secondary=NULL;
             }
-            wfxe->Samples.wValidBitsPerSample = wfxe->Format.wBitsPerSample;
+            wfxe.Samples.wValidBitsPerSample = wfxe.Format.wBitsPerSample;
 
             rc=IDirectSound8_CreateSoundBuffer(dso,&bufdesc,&secondary,NULL);
             ok(rc==DS_OK && secondary!=NULL,
@@ -892,8 +891,6 @@
                 ok(ref==0,"IDirectSoundBuffer_Release() has %d references, "
                    "should have 0\n",ref);
             }
-
-            free(wfxe);
           }
         }
 EXIT1:
