diff --git a/dlls/msvcrt/tests/cpp.c b/dlls/msvcrt/tests/cpp.c
--- a/dlls/msvcrt/tests/cpp.c
+++ b/dlls/msvcrt/tests/cpp.c
@@ -111,7 +111,7 @@
 /* Emulate a __thiscall */
 #ifdef __i386__
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 struct thiscall_thunk
 {
     BYTE pop_eax;    /* popl  %eax (ret addr) */
@@ -120,7 +120,7 @@
     BYTE push_eax;   /* pushl %eax */
     WORD jmp_edx;    /* jmp  *%edx */
 };
-#pragma pack(pop)
+#include "poppack.h"
 
 static void * (WINAPI *call_thiscall_func1)( void *func, void *this );
 static void * (WINAPI *call_thiscall_func2)( void *func, void *this, const void *a );
@@ -869,12 +869,15 @@
     return (void *)vtable[-1];
 }
 
-#ifndef _WIN64
-#define RTTI_REF(instance, name) (uintptr_t)&instance.name
+#ifdef __i386__
+#define DEFINE_RTTI_REF(type, name) type *name
+#define RTTI_REF(instance, name) &instance.name
+#define RTTI_REF_SIG0(instance, name, base) RTTI_REF(instance, name)
 #else
+#define DEFINE_RTTI_REF(type, name) unsigned name
 #define RTTI_REF(instance, name) FIELD_OFFSET(struct rtti_data, name)
+#define RTTI_REF_SIG0(instance, name, base) ((char*)&instance.name-base)
 #endif
-#define RVA(ptr,base) ((uintptr_t)ptr - base)
 /* Test RTTI functions */
 static void test_rtti(void)
 {
@@ -883,14 +886,14 @@
       unsigned int signature;
       int base_class_offset;
       unsigned int flags;
-      unsigned int type_descriptor;
-      unsigned int type_hierarchy;
-      unsigned int object_locator;
+      DEFINE_RTTI_REF(type_info, type_descriptor);
+      DEFINE_RTTI_REF(struct _rtti_object_hierarchy, type_hierarchy);
+      DEFINE_RTTI_REF(void, object_locator);
   } *obj_locator;
 
   struct _rtti_base_descriptor
   {
-    unsigned int type_descriptor;
+    DEFINE_RTTI_REF(type_info, type_descriptor);
     int num_base_classes;
     struct {
       int this_offset;
@@ -901,14 +904,14 @@
   };
 
   struct _rtti_base_array {
-    unsigned int bases[4];
+    DEFINE_RTTI_REF(struct _rtti_base_descriptor, bases[4]);
   };
 
   struct _rtti_object_hierarchy {
     unsigned int signature;
     unsigned int attributes;
     int array_len;
-    unsigned int base_classes;
+    DEFINE_RTTI_REF(struct _rtti_base_array, base_classes);
   };
 
   struct rtti_data
@@ -967,10 +970,8 @@
   exception e,b;
   void *casted;
   BOOL old_signature;
-#ifndef _WIN64
-  uintptr_t base = 0;
-#else
-  uintptr_t base = (uintptr_t)GetModuleHandleW(NULL);
+#ifndef __i386__
+  char *base = (char*)GetModuleHandleW(NULL);
 #endif
 
   if (!p__RTCastToVoid || !p__RTtypeid || !pexception_ctor || !pbad_typeid_ctor
@@ -1010,21 +1011,21 @@
 
   simple_class_sig0_rtti = simple_class_rtti;
   simple_class_sig0_rtti.object_locator.signature = 0;
-  simple_class_sig0_rtti.base_descriptor[0].type_descriptor = RVA(&simple_class_sig0_rtti.type_info[0], base);
-  simple_class_sig0_rtti.base_array.bases[0] = RVA(&simple_class_sig0_rtti.base_descriptor[0], base);
-  simple_class_sig0_rtti.object_hierarchy.base_classes = RVA(&simple_class_sig0_rtti.base_array, base);
-  simple_class_sig0_rtti.object_locator.type_descriptor = RVA(&simple_class_sig0_rtti.type_info[0], base);
-  simple_class_sig0_rtti.object_locator.type_hierarchy = RVA(&simple_class_sig0_rtti.object_hierarchy, base);
+  simple_class_sig0_rtti.base_descriptor[0].type_descriptor = RTTI_REF_SIG0(simple_class_sig0_rtti, type_info[0], base);
+  simple_class_sig0_rtti.base_array.bases[0] = RTTI_REF_SIG0(simple_class_sig0_rtti, base_descriptor[0], base);
+  simple_class_sig0_rtti.object_hierarchy.base_classes = RTTI_REF_SIG0(simple_class_sig0_rtti, base_array, base);
+  simple_class_sig0_rtti.object_locator.type_descriptor = RTTI_REF_SIG0(simple_class_sig0_rtti, type_info[0], base);
+  simple_class_sig0_rtti.object_locator.type_hierarchy = RTTI_REF_SIG0(simple_class_sig0_rtti, object_hierarchy, base);
 
   child_class_sig0_rtti = child_class_rtti;
   child_class_sig0_rtti.object_locator.signature = 0;
-  child_class_sig0_rtti.base_descriptor[0].type_descriptor = RVA(&child_class_sig0_rtti.type_info[1], base);
-  child_class_sig0_rtti.base_descriptor[1].type_descriptor = RVA(&child_class_sig0_rtti.type_info[0], base);
-  child_class_sig0_rtti.base_array.bases[0] = RVA(&child_class_sig0_rtti.base_descriptor[0], base);
-  child_class_sig0_rtti.base_array.bases[1] = RVA(&child_class_sig0_rtti.base_descriptor[1], base);
-  child_class_sig0_rtti.object_hierarchy.base_classes = RVA(&child_class_sig0_rtti.base_array, base);
-  child_class_sig0_rtti.object_locator.type_descriptor = RVA(&child_class_sig0_rtti.type_info[1], base);
-  child_class_sig0_rtti.object_locator.type_hierarchy = RVA(&child_class_sig0_rtti.object_hierarchy, base);
+  child_class_sig0_rtti.base_descriptor[0].type_descriptor = RTTI_REF_SIG0(child_class_sig0_rtti, type_info[1], base);
+  child_class_sig0_rtti.base_descriptor[1].type_descriptor = RTTI_REF_SIG0(child_class_sig0_rtti, type_info[0], base);
+  child_class_sig0_rtti.base_array.bases[0] = RTTI_REF_SIG0(child_class_sig0_rtti, base_descriptor[0], base);
+  child_class_sig0_rtti.base_array.bases[1] = RTTI_REF_SIG0(child_class_sig0_rtti, base_descriptor[1], base);
+  child_class_sig0_rtti.object_hierarchy.base_classes = RTTI_REF_SIG0(child_class_sig0_rtti, base_array, base);
+  child_class_sig0_rtti.object_locator.type_descriptor = RTTI_REF_SIG0(child_class_sig0_rtti, type_info[1], base);
+  child_class_sig0_rtti.object_locator.type_hierarchy = RTTI_REF_SIG0(child_class_sig0_rtti, object_hierarchy, base);
 
   ti = p__RTtypeid(&simple_class_sig0);
   ok (ti && !strcmp(ti->mangled, "simple_class"), "incorrect rtti data\n");
