diff --git a/dlls/d3dcompiler_43/tests/reflection.c b/dlls/d3dcompiler_43/tests/reflection.c
--- a/dlls/d3dcompiler_43/tests/reflection.c
+++ b/dlls/d3dcompiler_43/tests/reflection.c
@@ -107,10 +107,11 @@
     ok(!count, "Got unexpected count %lu.\n", count);
 
     /* check invalid cases */
-    if (D3D_COMPILER_VERSION >= 46)
-        expected = E_INVALIDARG;
-    else
-        expected = E_NOINTERFACE;
+#if D3D_COMPILER_VERSION >= 46
+    expected = E_INVALIDARG;
+#else
+    expected = E_NOINTERFACE;
+#endif
     hr = D3DReflect(test_reflection_blob, test_reflection_blob[6], &IID_ID3D10ShaderReflection, (void **)&ref10);
     ok(hr == expected || broken(hr == E_NOINTERFACE) /* Windows 8 */,
             "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
@@ -119,47 +120,34 @@
     ok(hr == expected || broken(hr == E_NOINTERFACE) /* Windows 8 */,
             "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
 
-    if (D3D_COMPILER_VERSION >= 43)
-        expected = D3DERR_INVALIDCALL;
-    else if (D3D_COMPILER_VERSION)
-        expected = E_INVALIDARG;
-    else
-        expected = E_FAIL;
     hr = D3DReflect(NULL, test_reflection_blob[6], &IID_ID3D10ShaderReflection1, (void **)&ref10_1);
-    ok(hr == expected, "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
+    ok(hr == D3DERR_INVALIDCALL, "Got unexpected hr %#lx.\n", hr);
 
     hr = D3DReflect(NULL, test_reflection_blob[6], &IID_ID3D11ShaderReflection, (void **)&ref11);
-    ok(hr == expected, "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
+    ok(hr == D3DERR_INVALIDCALL, "Got unexpected hr %#lx.\n", hr);
 
     /* returns different errors with different sizes */
     hr = D3DReflect(test_reflection_blob, 31, &IID_ID3D10ShaderReflection1, (void **)&ref10_1);
-    ok(hr == expected, "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
-
-    if (D3D_COMPILER_VERSION >= 46)
-        expected = D3DERR_INVALIDCALL;
-    else if (D3D_COMPILER_VERSION == 42)
-        expected = E_INVALIDARG;
-    else
-        expected = E_FAIL;
+    ok(hr == D3DERR_INVALIDCALL, "Got unexpected hr %#lx.\n", hr);
+
+#if D3D_COMPILER_VERSION >= 46
+    expected = D3DERR_INVALIDCALL;
+#else
+    expected = E_FAIL;
+#endif
     hr = D3DReflect(test_reflection_blob,  32, &IID_ID3D10ShaderReflection1, (void **)&ref10_1);
     ok(hr == expected, "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
 
     hr = D3DReflect(test_reflection_blob, test_reflection_blob[6]-1, &IID_ID3D10ShaderReflection1, (void **)&ref10_1);
     ok(hr == expected, "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
 
+    hr = D3DReflect(test_reflection_blob,  31, &IID_ID3D11ShaderReflection, (void **)&ref11);
+    ok(hr == D3DERR_INVALIDCALL, "Got unexpected hr %#lx.\n", hr);
+
     hr = D3DReflect(test_reflection_blob,  32, &IID_ID3D11ShaderReflection, (void **)&ref11);
     ok(hr == expected, "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
 
     hr = D3DReflect(test_reflection_blob,  test_reflection_blob[6]-1, &IID_ID3D11ShaderReflection, (void **)&ref11);
-    ok(hr == expected, "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
-
-    if (D3D_COMPILER_VERSION >= 43)
-        expected = D3DERR_INVALIDCALL;
-    else if (D3D_COMPILER_VERSION)
-        expected = E_INVALIDARG;
-    else
-        expected = D3DERR_INVALIDCALL;
-    hr = D3DReflect(test_reflection_blob,  31, &IID_ID3D11ShaderReflection, (void **)&ref11);
     ok(hr == expected, "Got unexpected hr %#lx, expected %#lx.\n", hr, expected);
 }
 #endif
@@ -788,9 +776,7 @@
     D3D_NAME expected;
     unsigned int i;
 #if D3D_COMPILER_VERSION
-#if D3D_COMPILER_VERSION >= 43
     UINT size_x, size_y, size_z, size_total;
-#endif
     UINT ret;
 #endif
 
@@ -934,7 +920,7 @@
             ok(!desc.Stream, "(%u): got unexpected Stream %u.\n", i, desc.Stream);
     }
 
-#if D3D_COMPILER_VERSION >= 43
+#if D3D_COMPILER_VERSION
     size_x = 0xdeadbeef;
     size_y = 0xdeadbeef;
     size_z = 0xdeadbeef;
@@ -1084,7 +1070,7 @@
     for (i = 0; i < ARRAY_SIZE(tests); ++i)
     {
         hr = call_reflect(tests[i].blob, tests[i].blob[6], &IID_ID3D11ShaderReflection, (void **)&ref11);
-        if (D3D_COMPILER_VERSION < 43)
+        if (!D3D_COMPILER_VERSION)
         {
             todo_wine ok(hr == E_INVALIDARG, "%u: Got unexpected hr %#lx.\n", i, hr);
             if (SUCCEEDED(hr))
@@ -1278,13 +1264,10 @@
     if (FAILED(hr))
         return;
     size_total = ref11->lpVtbl->GetThreadGroupSize(ref11, &size_x, &size_y, &size_z);
-    if (D3D_COMPILER_VERSION >= 43)
-    {
-        ok(size_x == 16, "Unexpected size %u.\n", size_x);
-        ok(size_y == 8, "Unexpected size %u.\n", size_y);
-        ok(size_z == 4, "Unexpected size %u.\n", size_z);
-        ok(size_total == size_x * size_y * size_z, "Unexpected size %u.\n", size_total);
-    }
+    ok(size_x == 16, "Unexpected size %u.\n", size_x);
+    ok(size_y == 8, "Unexpected size %u.\n", size_y);
+    ok(size_z == 4, "Unexpected size %u.\n", size_z);
+    ok(size_total == size_x * size_y * size_z, "Unexpected size %u.\n", size_total);
 
     ref11->lpVtbl->Release(ref11);
 }
