diff --git a/dlls/user32/tests/winstation.c b/dlls/user32/tests/winstation.c
--- a/dlls/user32/tests/winstation.c
+++ b/dlls/user32/tests/winstation.c
@@ -172,6 +172,10 @@
     else if (le == ERROR_ACCESS_DENIED)
         win_skip( "Not enough privileges for CreateWindowStation\n" );
 
+    w2 = OpenWindowStationA("winsta0", TRUE, 0 );
+    ok( !w2, "got non-NULL.\n" );
+    ok( GetLastError() == ERROR_ACCESS_DENIED, "got %ld.\n", GetLastError() );
+
     w2 = OpenWindowStationA("winsta0", TRUE, WINSTA_ALL_ACCESS );
     ok( w2 != 0, "OpenWindowStation failed\n" );
     ok( w2 != w1, "OpenWindowStation returned default handle\n" );
@@ -858,20 +862,30 @@
     return 0;
 }
 
-/* try to make sure pending X events have been processed before continuing */
-static void flush_events(void)
-{
-    MSG msg;
-    int diff = 200;
-    int min_timeout = 100;
-    DWORD time = GetTickCount() + diff;
-
-    while (diff > 0)
-    {
-        if (MsgWaitForMultipleObjects( 0, NULL, FALSE, min_timeout, QS_ALLINPUT ) == WAIT_TIMEOUT) break;
-        while (PeekMessageA( &msg, 0, 0, 0, PM_REMOVE )) DispatchMessageA( &msg );
-        diff = time - GetTickCount();
-    }
+static DWORD set_foreground(HWND hwnd)
+{
+    HWND hwnd_fore;
+    DWORD set_id, fore_id, ret;
+    char win_text[1024];
+
+    hwnd_fore = GetForegroundWindow();
+    GetWindowTextA(hwnd_fore, win_text, 1024);
+    set_id = GetWindowThreadProcessId(hwnd, NULL);
+    fore_id = GetWindowThreadProcessId(hwnd_fore, NULL);
+    trace("\"%s\" %p %08lx hwnd %p %08lx\n", win_text, hwnd_fore, fore_id, hwnd, set_id);
+    ret = AttachThreadInput(set_id, fore_id, TRUE);
+    trace("AttachThreadInput returned %08lx\n", ret);
+    ret = ShowWindow(hwnd, SW_SHOWNORMAL);
+    trace("ShowWindow returned %08lx\n", ret);
+    ret = SetWindowPos(hwnd, HWND_TOPMOST, 0,0,0,0, SWP_NOSIZE|SWP_NOMOVE);
+    trace("set topmost returned %08lx\n", ret);
+    ret = SetWindowPos(hwnd, HWND_NOTOPMOST, 0,0,0,0, SWP_NOSIZE|SWP_NOMOVE);
+    trace("set notopmost returned %08lx\n", ret);
+    ret = SetForegroundWindow(hwnd);
+    trace("SetForegroundWindow returned %08lx\n", ret);
+    Sleep(250);
+    AttachThreadInput(set_id, fore_id, FALSE);
+    return ret;
 }
 
 static void test_foregroundwindow(void)
@@ -945,21 +959,18 @@
         for (thread_desk_id = 0; thread_desk_id < DESKTOPS; thread_desk_id++)
             for (input_desk_id = 0; input_desk_id < DESKTOPS; input_desk_id++)
             {
-                winetest_push_context("thread %d input %d hwnd %d",
+                trace("testing thread_desk %d input_desk %d hwnd %d\n",
                         thread_desk_id, input_desk_id, hwnd_id);
                 hwnd_test = hwnds[hwnd_id];
                 ret = SetThreadDesktop(hdesks[thread_desk_id]);
                 ok(ret, "set thread desktop failed!\n");
                 ret = SwitchDesktop(hdesks[input_desk_id]);
                 ok(ret, "switch desktop failed!\n");
-                SetForegroundWindow(partners[0]);
-                flush_events();
-                SetForegroundWindow(partners[1]);
-                flush_events();
+                set_foreground(partners[0]);
+                set_foreground(partners[1]);
                 hwnd = GetForegroundWindow();
                 ok(hwnd != hwnd_test, "unexpected foreground window %p\n", hwnd);
-                ret = SetForegroundWindow(hwnd_test);
-                flush_events();
+                ret = set_foreground(hwnd_test);
                 hwnd = GetForegroundWindow();
                 GetWindowTextA(hwnd, win_text, 1024);
                 trace("hwnd %p name %s\n", hwnd, win_text);
@@ -968,7 +979,8 @@
                     if (input_desk_id == thread_desk_id)
                     {
                         ok(ret, "SetForegroundWindow failed!\n");
-                        ok(hwnd == hwnd_test , "unexpected foreground window %p\n", hwnd);
+                        todo_wine_if (!hwnd)
+                            ok(hwnd == hwnd_test , "unexpected foreground window %p\n", hwnd);
                     }
                     else
                     {
@@ -981,15 +993,16 @@
                     if (input_desk_id == thread_desk_id)
                     {
                         ok(!ret, "SetForegroundWindow should fail!\n");
-                        ok(hwnd == partners[input_desk_id] , "unexpected foreground window %p\n", hwnd);
+                        todo_wine_if (!hwnd)
+                            ok(hwnd == partners[input_desk_id] , "unexpected foreground window %p\n", hwnd);
                     }
                     else
                     {
                         todo_wine ok(!ret, "SetForegroundWindow should fail!\n");
-                        todo_wine ok(hwnd == 0, "unexpected foreground window %p\n", hwnd);
+                        todo_wine_if (hwnd)
+                            ok(hwnd == 0, "unexpected foreground window %p\n", hwnd);
                     }
                 }
-                winetest_pop_context();
             }
 
     /* Clean up */
@@ -1072,7 +1085,9 @@
     ret = CreateProcessA(argv[0], buffer, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
     ok(ret, "CreateProcessA failed, error %lu.\n", GetLastError());
 
-    wait_child_process(&pi);
+    wait_child_process(pi.hProcess);
+    CloseHandle(pi.hThread);
+    CloseHandle(pi.hProcess);
     CloseDesktop(desktop);
     CloseWindowStation(winstation);
     SetProcessWindowStation(old_winstation);
