diff --git a/dlls/mshtml/htmlevent.c b/dlls/mshtml/htmlevent.c
--- a/dlls/mshtml/htmlevent.c
+++ b/dlls/mshtml/htmlevent.c
@@ -161,8 +161,6 @@
         EVENT_BUBBLES | EVENT_CANCELABLE},
     {L"input",             EVENT_TYPE_EVENT,     DISPID_UNKNOWN,
         EVENT_DEFAULTLISTENER | EVENT_BUBBLES},
-    {L"invalid",           EVENT_TYPE_EVENT,     DISPID_EVPROP_INVALID,
-        EVENT_BIND_TO_TARGET | EVENT_CANCELABLE},
     {L"keydown",           EVENT_TYPE_KEYBOARD,  DISPID_EVMETH_ONKEYDOWN,
         EVENT_DEFAULTLISTENER | EVENT_HASDEFAULTHANDLERS | EVENT_BUBBLES | EVENT_CANCELABLE },
     {L"keypress",          EVENT_TYPE_KEYBOARD,  DISPID_EVMETH_ONKEYPRESS,
@@ -1811,7 +1809,7 @@
 };
 
 dispex_static_data_t MSEventObj_dispex = {
-    .id         = OBJID_MSEventObj,
+    .id         = PROT_MSEventObj,
     .vtbl       = &HTMLEventObj_dispex_vtbl,
     .disp_tid   = DispCEventObj_tid,
     .iface_tids = MSEventObj_iface_tids,
@@ -3584,7 +3582,7 @@
 };
 
 dispex_static_data_t Event_dispex = {
-    .id         = OBJID_Event,
+    .id         = PROT_Event,
     .vtbl       = &DOMEvent_dispex_vtbl,
     .disp_tid   = DispDOMEvent_tid,
     .iface_tids = Event_iface_tids,
@@ -3604,8 +3602,8 @@
 };
 
 dispex_static_data_t UIEvent_dispex = {
-    .id           = OBJID_UIEvent,
-    .prototype_id = OBJID_Event,
+    .id           = PROT_UIEvent,
+    .prototype_id = PROT_Event,
     .vtbl         = &DOMUIEvent_dispex_vtbl,
     .disp_tid     = DispDOMUIEvent_tid,
     .iface_tids   = UIEvent_iface_tids,
@@ -3626,8 +3624,8 @@
 };
 
 dispex_static_data_t MouseEvent_dispex = {
-    .id           = OBJID_MouseEvent,
-    .prototype_id = OBJID_UIEvent,
+    .id           = PROT_MouseEvent,
+    .prototype_id = PROT_UIEvent,
     .vtbl         = &DOMMouseEvent_dispex_vtbl,
     .disp_tid     = DispDOMMouseEvent_tid,
     .iface_tids   = MouseEvent_iface_tids,
@@ -3656,8 +3654,8 @@
 };
 
 dispex_static_data_t KeyboardEvent_dispex = {
-    .id           = OBJID_KeyboardEvent,
-    .prototype_id = OBJID_UIEvent,
+    .id           = PROT_KeyboardEvent,
+    .prototype_id = PROT_UIEvent,
     .vtbl         = &DOMKeyboardEvent_dispex_vtbl,
     .disp_tid     = DispDOMKeyboardEvent_tid,
     .iface_tids   = KeyboardEvent_iface_tids,
@@ -3677,8 +3675,8 @@
 };
 
 dispex_static_data_t PageTransitionEvent_dispex = {
-    .id              = OBJID_PageTransitionEvent,
-    .prototype_id    = OBJID_Event,
+    .id              = PROT_PageTransitionEvent,
+    .prototype_id    = PROT_Event,
     .vtbl            = &DOMPageTransitionEvent_dispex_vtbl,
     .disp_tid        = DispDOMEvent_tid,
     .iface_tids      = Event_iface_tids,
@@ -3700,8 +3698,8 @@
 };
 
 dispex_static_data_t CustomEvent_dispex = {
-    .id           = OBJID_CustomEvent,
-    .prototype_id = OBJID_Event,
+    .id           = PROT_CustomEvent,
+    .prototype_id = PROT_Event,
     .vtbl         = &DOMCustomEvent_dispex_vtbl,
     .disp_tid     = DispDOMCustomEvent_tid,
     .iface_tids   = CustomEvent_iface_tids,
@@ -3715,8 +3713,8 @@
 };
 
 dispex_static_data_t MessageEvent_dispex = {
-    .id           = OBJID_MessageEvent,
-    .prototype_id = OBJID_Event,
+    .id           = PROT_MessageEvent,
+    .prototype_id = PROT_Event,
     .vtbl         = &DOMMessageEvent_dispex_vtbl,
     .disp_tid     = DispDOMMessageEvent_tid,
     .iface_tids   = Event_iface_tids,
@@ -3737,8 +3735,8 @@
 };
 
 dispex_static_data_t ProgressEvent_dispex = {
-    .id              = OBJID_ProgressEvent,
-    .prototype_id    = OBJID_Event,
+    .id              = PROT_ProgressEvent,
+    .prototype_id    = PROT_Event,
     .vtbl            = &DOMProgressEvent_dispex_vtbl,
     .disp_tid        = DispDOMProgressEvent_tid,
     .iface_tids      = ProgressEvent_iface_tids,
@@ -3759,8 +3757,8 @@
 };
 
 dispex_static_data_t StorageEvent_dispex = {
-    .id           = OBJID_StorageEvent,
-    .prototype_id = OBJID_Event,
+    .id           = PROT_StorageEvent,
+    .prototype_id = PROT_Event,
     .vtbl         = &DOMStorageEvent_dispex_vtbl,
     .disp_tid     = DispDOMStorageEvent_tid,
     .iface_tids   = StorageEvent_iface_tids,
@@ -4966,9 +4964,8 @@
     unsigned i;
     HRESULT hres;
 
-    if(doc->event_vector)
-        memset(doc->event_vector, 0, EVENTID_LAST * sizeof(BOOL));
-    else if(!(doc->event_vector = calloc(EVENTID_LAST, sizeof(BOOL))))
+    doc->event_vector = calloc(EVENTID_LAST, sizeof(BOOL));
+    if(!doc->event_vector)
         return E_OUTOFMEMORY;
 
     init_nsevents(doc);
