diff --git a/dlls/ntdll/unix/debug.c b/dlls/ntdll/unix/debug.c
--- a/dlls/ntdll/unix/debug.c
+++ b/dlls/ntdll/unix/debug.c
@@ -33,6 +33,7 @@
 #include <sys/types.h>
 #include <sys/stat.h>
 #include <unistd.h>
+#include <errno.h>
 
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
@@ -44,7 +45,9 @@
 
 WINE_DECLARE_DEBUG_CHANNEL(pid);
 WINE_DECLARE_DEBUG_CHANNEL(timestamp);
+WINE_DECLARE_DEBUG_CHANNEL(microsecs);
 WINE_DEFAULT_DEBUG_CHANNEL(ntdll);
+WINE_DECLARE_DEBUG_CHANNEL(ftracelog);
 
 struct debug_info
 {
@@ -282,6 +285,60 @@
     return write( 2, params->str, params->len );
 }
 
+static void __wine_dbg_ftrace_write( const char *str, unsigned int str_len )
+{
+    static int ftrace_fd = -1;
+
+    if (ftrace_fd == -1)
+    {
+        int expected = -1;
+        const char *fn;
+        int fd;
+
+        if (!(fn = getenv( "WINE_FTRACE_FILE" ))) fn = "/sys/kernel/tracing/trace_marker";
+        if ((fd = open( fn, O_WRONLY )) == -1)
+        {
+            MESSAGE( "wine: error opening ftrace file: %s.\n", strerror(errno) );
+            ftrace_fd = -2;
+            return;
+        }
+        if (!__atomic_compare_exchange_n( &ftrace_fd, &expected, fd, 0, __ATOMIC_SEQ_CST, __ATOMIC_SEQ_CST ))
+            close( fd );
+        else
+            MESSAGE( "wine: ftrace initialized.\n" );
+    }
+
+    if (ftrace_fd == -2) return;
+    write( ftrace_fd, str, str_len );
+}
+
+unsigned int WINAPI __wine_dbg_ftrace( char *str, unsigned int str_size, unsigned int ctx )
+{
+    static unsigned int curr_ctx;
+    unsigned int str_len;
+    char ctx_str[64];
+    int ctx_len;
+
+    if (ctx == ~0u) ctx_len = 0;
+    else if (ctx) ctx_len = sprintf( ctx_str, " (end_ctx=%u)", ctx );
+    else
+    {
+        ctx = __atomic_add_fetch( &curr_ctx, 1, __ATOMIC_SEQ_CST );
+        ctx_len = sprintf( ctx_str, " (begin_ctx=%u)", ctx );
+    }
+
+    str_len = strlen(str);
+    if (ctx_len > 0)
+    {
+        if (str_size < ctx_len) return ~0u;
+        if (str_len + ctx_len > str_size) str_len = str_size - ctx_len;
+        memcpy( &str[str_len], ctx_str, ctx_len );
+        str_len += ctx_len;
+    }
+    __wine_dbg_ftrace_write( str, str_len );
+    return ctx;
+}
+
 #ifdef _WIN64
 /***********************************************************************
  *		wow64_wine_dbg_write
@@ -311,6 +368,7 @@
     {
         ret += append_output( info, str, end + 1 - str );
         write( 2, info->output, info->out_pos );
+        if (TRACE_ON(ftracelog)) __wine_dbg_ftrace_write( info->output, info->out_pos );
         info->out_pos = 0;
         str = end + 1;
     }
@@ -335,13 +393,20 @@
 
     if (init_done)
     {
-        if (TRACE_ON(timestamp))
+        if (TRACE_ON(microsecs))
+        {
+            LARGE_INTEGER counter, frequency, microsecs;
+            NtQueryPerformanceCounter(&counter, &frequency);
+            microsecs.QuadPart = counter.QuadPart * 1000000 / frequency.QuadPart;
+            pos += sprintf( pos, "%3u.%06u:", (unsigned int)(microsecs.QuadPart / 1000000), (unsigned int)(microsecs.QuadPart % 1000000) );
+        }
+        else if (TRACE_ON(timestamp))
         {
             UINT ticks = NtGetTickCount();
             pos += snprintf( pos, sizeof(info->output) - (pos - info->output), "%3u.%03u:", ticks / 1000, ticks % 1000 );
         }
-        if (TRACE_ON(pid)) pos += snprintf( pos, sizeof(info->output) - (pos - info->output), "%04x:", GetCurrentProcessId() );
-        pos += snprintf( pos, sizeof(info->output) - (pos - info->output), "%04x:", GetCurrentThreadId() );
+        if (TRACE_ON(pid)) pos += snprintf( pos, sizeof(info->output) - (pos - info->output), "%04x:", (UINT)GetCurrentProcessId() );
+        pos += snprintf( pos, sizeof(info->output) - (pos - info->output), "%04x:", (UINT)GetCurrentThreadId() );
     }
     if (function && cls < ARRAY_SIZE( classes ))
         pos += snprintf( pos, sizeof(info->output) - (pos - info->output), "%s:%s:%s ",
@@ -378,7 +443,7 @@
                                 void *outbuf, ULONG outbuf_len, ULONG *size )
 {
     FIXME( "code %u, inbuf %p, inbuf_len %u, outbuf %p, outbuf_len %u, size %p\n",
-           code, inbuf, inbuf_len, outbuf, outbuf_len, size );
+           (int)code, inbuf, (int)inbuf_len, outbuf, (int)outbuf_len, size );
     return STATUS_SUCCESS;
 }
 
@@ -388,7 +453,7 @@
  */
 NTSTATUS WINAPI NtSetDebugFilterState( ULONG component_id, ULONG level, BOOLEAN state )
 {
-    FIXME( "component_id %#x, level %u, state %#x stub.\n", component_id, level, state );
+    FIXME( "component_id %#x, level %u, state %#x stub.\n", (int)component_id, (int)level, state );
 
     return STATUS_SUCCESS;
 }
