diff --git a/dlls/quartz/tests/filtergraph.c b/dlls/quartz/tests/filtergraph.c
--- a/dlls/quartz/tests/filtergraph.c
+++ b/dlls/quartz/tests/filtergraph.c
@@ -862,14 +862,6 @@
     HANDLE on_input_full;
 };
 
-struct testenumpins
-{
-    IEnumPins IEnumPins_iface;
-    LONG ref;
-    struct testfilter *filter;
-    unsigned int enum_idx;
-};
-
 struct testfilter
 {
     IBaseFilter IBaseFilter_iface;
@@ -878,8 +870,9 @@
     WCHAR *name;
     IReferenceClock *clock;
 
+    IEnumPins IEnumPins_iface;
     struct testpin *pins;
-    unsigned int pin_count;
+    unsigned int pin_count, enum_idx;
 
     FILTER_STATE state;
     REFERENCE_TIME start_time;
@@ -1271,9 +1264,9 @@
     pin->type_count = type_count;
 }
 
-static inline struct testenumpins *impl_from_IEnumPins(IEnumPins *iface)
-{
-    return CONTAINING_RECORD(iface, struct testenumpins, IEnumPins_iface);
+static inline struct testfilter *impl_from_IEnumPins(IEnumPins *iface)
+{
+    return CONTAINING_RECORD(iface, struct testfilter, IEnumPins_iface);
 }
 
 static HRESULT WINAPI testenumpins_QueryInterface(IEnumPins *iface, REFIID iid, void **out)
@@ -1284,37 +1277,33 @@
 
 static ULONG WINAPI testenumpins_AddRef(IEnumPins * iface)
 {
-    struct testenumpins *enumpins = impl_from_IEnumPins(iface);
-    return InterlockedIncrement(&enumpins->ref);
+    struct testfilter *filter = impl_from_IEnumPins(iface);
+    return InterlockedIncrement(&filter->ref);
 }
 
 static ULONG WINAPI testenumpins_Release(IEnumPins * iface)
 {
-    struct testenumpins *enumpins = impl_from_IEnumPins(iface);
-    ULONG ref = InterlockedIncrement(&enumpins->ref);
-
-    if (!ref)
-        free(enumpins);
-    return ref;
+    struct testfilter *filter = impl_from_IEnumPins(iface);
+    return InterlockedDecrement(&filter->ref);
 }
 
 static HRESULT WINAPI testenumpins_Next(IEnumPins *iface, ULONG count, IPin **out, ULONG *fetched)
 {
-    struct testenumpins *enumpins = impl_from_IEnumPins(iface);
+    struct testfilter *filter = impl_from_IEnumPins(iface);
     unsigned int i;
 
     for (i = 0; i < count; ++i)
     {
-        if (enumpins->enum_idx + i >= enumpins->filter->pin_count)
+        if (filter->enum_idx + i >= filter->pin_count)
             break;
 
-        out[i] = &enumpins->filter->pins[enumpins->enum_idx + i].IPin_iface;
+        out[i] = &filter->pins[filter->enum_idx + i].IPin_iface;
         IPin_AddRef(out[i]);
     }
 
     if (fetched)
         *fetched = i;
-    enumpins->enum_idx += i;
+    filter->enum_idx += i;
 
     return (i == count) ? S_OK : S_FALSE;
 }
@@ -1327,8 +1316,8 @@
 
 static HRESULT WINAPI testenumpins_Reset(IEnumPins *iface)
 {
-    struct testenumpins *enumpins = impl_from_IEnumPins(iface);
-    enumpins->enum_idx = 0;
+    struct testfilter *filter = impl_from_IEnumPins(iface);
+    filter->enum_idx = 0;
     return S_OK;
 }
 
@@ -1513,17 +1502,11 @@
 static HRESULT WINAPI testfilter_EnumPins(IBaseFilter *iface, IEnumPins **out)
 {
     struct testfilter *filter = impl_from_IBaseFilter(iface);
-    struct testenumpins *enumpins;
-
     if (winetest_debug > 1) trace("%p->EnumPins()\n", filter);
 
-    enumpins = malloc(sizeof(*enumpins));
-    enumpins->IEnumPins_iface.lpVtbl = &testenumpins_vtbl;
-    enumpins->ref = 1;
-    enumpins->enum_idx = 0;
-    enumpins->filter = filter;
-
-    *out = &enumpins->IEnumPins_iface;
+    *out = &filter->IEnumPins_iface;
+    IEnumPins_AddRef(*out);
+    filter->enum_idx = 0;
     return S_OK;
 }
 
@@ -2067,6 +2050,7 @@
 
     memset(filter, 0, sizeof(*filter));
     filter->IBaseFilter_iface.lpVtbl = &testfilter_vtbl;
+    filter->IEnumPins_iface.lpVtbl = &testenumpins_vtbl;
     filter->ref = 1;
     filter->pins = pins;
     filter->pin_count = pin_count;
@@ -3123,7 +3107,7 @@
 
     IFilterGraph2 *graph = create_graph();
     IMediaControl *control;
-    AM_MEDIA_TYPE mt = {0};
+    AM_MEDIA_TYPE mt;
     HRESULT hr;
     ULONG ref;
 
