diff --git a/dlls/winegstreamer/wg_format.c b/dlls/winegstreamer/wg_format.c
--- a/dlls/winegstreamer/wg_format.c
+++ b/dlls/winegstreamer/wg_format.c
@@ -151,6 +151,8 @@
             return WG_VIDEO_FORMAT_AYUV;
         case GST_VIDEO_FORMAT_I420:
             return WG_VIDEO_FORMAT_I420;
+        case GST_VIDEO_FORMAT_I420_10LE:
+            return WG_VIDEO_FORMAT_P010_10LE;
         case GST_VIDEO_FORMAT_NV12:
             return WG_VIDEO_FORMAT_NV12;
         case GST_VIDEO_FORMAT_P010_10LE:
@@ -352,45 +354,6 @@
     }
 
     format->major_type = WG_MAJOR_TYPE_VIDEO_CINEPAK;
-    format->u.video.width = width;
-    format->u.video.height = height;
-    format->u.video.fps_n = fps_n;
-    format->u.video.fps_d = fps_d;
-}
-
-static void wg_format_from_caps_video_indeo(struct wg_format *format, const GstCaps *caps)
-{
-    const GstStructure *structure = gst_caps_get_structure(caps, 0);
-    gint version, width, height, fps_n, fps_d;
-
-    if (!gst_structure_get_int(structure, "indeoversion", &version))
-    {
-        GST_WARNING("Missing \"indeoversion\" value.");
-        return;
-    }
-    if (version != 5)
-    {
-        GST_FIXME("Unknown version %d.", version);
-        return;
-    }
-    if (!gst_structure_get_int(structure, "width", &width))
-    {
-        GST_WARNING("Missing \"width\" value.");
-        return;
-    }
-    if (!gst_structure_get_int(structure, "height", &height))
-    {
-        GST_WARNING("Missing \"height\" value.");
-        return;
-    }
-    if (!gst_structure_get_fraction(structure, "framerate", &fps_n, &fps_d))
-    {
-        fps_n = 0;
-        fps_d = 1;
-    }
-
-    format->major_type = WG_MAJOR_TYPE_VIDEO_INDEO;
-    format->u.video.version = version;
     format->u.video.width = width;
     format->u.video.height = height;
     format->u.video.fps_n = fps_n;
@@ -624,15 +587,19 @@
     {
         GstAudioInfo info;
 
-        if (gst_audio_info_from_caps(&info, caps))
+        if (gst_caps_is_fixed(caps) && gst_audio_info_from_caps(&info, caps))
             wg_format_from_audio_info(format, &info);
+        else
+            GST_WARNING("Unable to get audio info from caps");
     }
     else if (!strcmp(name, "video/x-raw"))
     {
         GstVideoInfo info;
 
-        if (gst_video_info_from_caps(&info, caps))
+        if (gst_caps_is_fixed(caps) && gst_video_info_from_caps(&info, caps))
             wg_format_from_video_info(format, &info);
+        else
+            GST_WARNING("Unable to get video info from caps");
     }
     else if (!strcmp(name, "audio/mpeg") && gst_structure_get_boolean(structure, "parsed", &parsed) && parsed)
     {
@@ -651,10 +618,6 @@
     else if (!strcmp(name, "video/x-cinepak"))
     {
         wg_format_from_caps_video_cinepak(format, caps);
-    }
-    else if (!strcmp(name, "video/x-indeo"))
-    {
-        wg_format_from_caps_video_indeo(format, caps);
     }
     else if (!strcmp(name, "video/x-wmv"))
     {
@@ -879,10 +842,20 @@
     GstBuffer *buffer;
     GstCaps *caps;
 
-    if (!(caps = gst_caps_new_empty_simple("audio/x-wma")))
-        return NULL;
-    if (format->u.audio.version)
-        gst_caps_set_simple(caps, "wmaversion", G_TYPE_INT, format->u.audio.version, NULL);
+    if (format->u.audio.is_xma)
+    {
+        if (!(caps = gst_caps_new_empty_simple("audio/x-xma")))
+            return NULL;
+        if (format->u.audio.version)
+            gst_caps_set_simple(caps, "xmaversion", G_TYPE_INT, format->u.audio.version, NULL);
+    }
+    else
+    {
+        if (!(caps = gst_caps_new_empty_simple("audio/x-wma")))
+            return NULL;
+        if (format->u.audio.version)
+            gst_caps_set_simple(caps, "wmaversion", G_TYPE_INT, format->u.audio.version, NULL);
+    }
 
     if (format->u.audio.bitrate)
         gst_caps_set_simple(caps, "bitrate", G_TYPE_INT, format->u.audio.bitrate, NULL);
@@ -1136,6 +1109,7 @@
         case WG_MAJOR_TYPE_AUDIO_MPEG4:
         case WG_MAJOR_TYPE_AUDIO_WMA:
         case WG_MAJOR_TYPE_VIDEO_H264:
+        case WG_MAJOR_TYPE_VIDEO_INDEO:
         case WG_MAJOR_TYPE_VIDEO_MPEG1:
             GST_FIXME("Format %u not implemented!", a->major_type);
             /* fallthrough */
@@ -1159,12 +1133,6 @@
             return a->u.video.width == b->u.video.width
                     && a->u.video.height == b->u.video.height;
 
-        case WG_MAJOR_TYPE_VIDEO_INDEO:
-            /* Do not compare FPS. */
-            return a->u.video.version == b->u.video.version
-                    && a->u.video.width == b->u.video.width
-                    && a->u.video.height == b->u.video.height;
-
         case WG_MAJOR_TYPE_VIDEO_WMV:
             /* Do not compare FPS. */
             return a->u.video.format == b->u.video.format
