diff --git a/dlls/d3dx9_36/skin.c b/dlls/d3dx9_36/skin.c
--- a/dlls/d3dx9_36/skin.c
+++ b/dlls/d3dx9_36/skin.c
@@ -2,7 +2,6 @@
  * Skin Info operations specific to D3DX9.
  *
  * Copyright (C) 2011 Dylan Smith
- * Copyright (C) 2013 Christian Costa
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -40,7 +39,7 @@
 
     DWORD fvf;
     D3DVERTEXELEMENT9 vertex_declaration[MAX_FVF_DECL_SIZE];
-    DWORD vertex_count;
+    DWORD num_vertices;
     DWORD num_bones;
     struct bone *bones;
 };
@@ -302,7 +301,7 @@
 
     TRACE("iface %p, skin_info %p.\n", iface, skin_info);
 
-    if (FAILED(hr = D3DXCreateSkinInfo(skin->vertex_count, skin->vertex_declaration, skin->num_bones, skin_info)))
+    if (FAILED(hr = D3DXCreateSkinInfo(skin->num_vertices, skin->vertex_declaration, skin->num_bones, skin_info)))
         return hr;
 
     for (i = 0; i < skin->num_bones; ++i)
@@ -397,97 +396,10 @@
 static HRESULT WINAPI d3dx9_skin_info_UpdateSkinnedMesh(ID3DXSkinInfo *iface, const D3DXMATRIX *bone_transforms,
         const D3DXMATRIX *bone_inv_transpose_transforms, const void *src_vertices, void *dst_vertices)
 {
-    struct d3dx9_skin_info *skin = impl_from_ID3DXSkinInfo(iface);
-    DWORD vertex_size = D3DXGetDeclVertexSize(skin->vertex_declaration, 0);
-
-    TRACE("iface %p, bone_transforms %p, bone_inv_transpose_transforms %p, src_vertices %p, dst_vertices %p.\n",
-            skin, bone_transforms, bone_inv_transpose_transforms, src_vertices, dst_vertices);
-
-    for (unsigned int i = 0; i < skin->vertex_count; ++i)
-    {
-        for (const D3DVERTEXELEMENT9 *element = skin->vertex_declaration; element->Stream != 0xff; ++element)
-        {
-            const void *src_element = (uint8_t *)src_vertices + (i * vertex_size) + element->Offset;
-            void *dst_element = (uint8_t *)dst_vertices + (i * vertex_size) + element->Offset;
-            unsigned int element_size = d3dx_decltype_size[element->Type];
-
-            switch (element->Usage)
-            {
-                case D3DDECLUSAGE_POSITION:
-                case D3DDECLUSAGE_NORMAL:
-                    memset(dst_element, 0, element_size);
-                    break;
-
-                default:
-                    FIXME("Unhandled usage %#x.\n", element->Usage);
-                    /* fall through */
-                case D3DDECLUSAGE_TEXCOORD:
-                    memcpy(dst_element, src_element, element_size);
-                    break;
-            }
-        }
-    }
-
-    for (const D3DVERTEXELEMENT9 *element = skin->vertex_declaration; element->Stream != 0xff; ++element)
-    {
-        if (element->Usage != D3DDECLUSAGE_POSITION && element->Usage != D3DDECLUSAGE_NORMAL)
-            continue;
-
-        for (unsigned int i = 0; i < skin->num_bones; ++i)
-        {
-            const struct bone *bone = &skin->bones[i];
-
-            for (unsigned int j = 0; j < bone->num_influences; ++j)
-            {
-                unsigned int vertex_idx = bone->vertices[j];
-                float weight = bone->weights[j];
-
-                const void *src_element = (uint8_t *)src_vertices + (vertex_idx * vertex_size) + element->Offset;
-                void *dst_element = (uint8_t *)dst_vertices + (vertex_idx * vertex_size) + element->Offset;
-                D3DXVECTOR3 *dst_vec3 = dst_element;
-
-                switch (element->Usage)
-                {
-                    case D3DDECLUSAGE_POSITION:
-                        if (element->Type == D3DDECLTYPE_FLOAT3)
-                        {
-                            D3DXVECTOR3 position;
-
-                            D3DXVec3TransformCoord(&position, src_element, &bone_transforms[i]);
-                            dst_vec3->x += weight * position.x;
-                            dst_vec3->y += weight * position.y;
-                            dst_vec3->z += weight * position.z;
-                        }
-                        else
-                        {
-                            FIXME("Unhandled position type %#x.\n", element->Type);
-                            return E_NOTIMPL;
-                        }
-                        break;
-
-                    case D3DDECLUSAGE_NORMAL:
-                    {
-                        D3DXVECTOR3 normal;
-
-                        if (element->Type != D3DDECLTYPE_FLOAT3)
-                        {
-                            FIXME("Unhandled normal type %#x.\n", element->Type);
-                            return E_NOTIMPL;
-                        }
-
-                        D3DXVec3TransformNormal(&normal, src_element, bone_inv_transpose_transforms
-                                ? &bone_inv_transpose_transforms[i] : &bone_transforms[i]);
-                        dst_vec3->x += weight * normal.x;
-                        dst_vec3->y += weight * normal.y;
-                        dst_vec3->z += weight * normal.z;
-                        break;
-                    }
-                }
-            }
-        }
-    }
-
-    return D3D_OK;
+    FIXME("iface %p, bone_transforms %p, bone_inv_transpose_transforms %p, src_vertices %p, dst_vertices %p stub!\n",
+            iface, bone_transforms, bone_inv_transpose_transforms, src_vertices, dst_vertices);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI d3dx9_skin_info_ConvertToBlendedMesh(ID3DXSkinInfo *iface, ID3DXMesh *mesh_in,
@@ -567,7 +479,7 @@
 
     object->ID3DXSkinInfo_iface.lpVtbl = &d3dx9_skin_info_vtbl;
     object->ref = 1;
-    object->vertex_count = vertex_count;
+    object->num_vertices = vertex_count;
     object->num_bones = bone_count;
     object->vertex_declaration[0] = empty_declaration;
     object->fvf = 0;
