diff --git a/dlls/wineoss.drv/oss.c b/dlls/wineoss.drv/oss.c
--- a/dlls/wineoss.drv/oss.c
+++ b/dlls/wineoss.drv/oss.c
@@ -476,12 +476,14 @@
     return ret;
 }
 
-static HRESULT setup_oss_device(AUDCLNT_SHAREMODE share, int fd, const WAVEFORMATEX *fmt)
+static HRESULT setup_oss_device(AUDCLNT_SHAREMODE share, int fd,
+                                const WAVEFORMATEX *fmt, WAVEFORMATEXTENSIBLE *out)
 {
     const WAVEFORMATEXTENSIBLE *fmtex = (const WAVEFORMATEXTENSIBLE *)fmt;
     int tmp, oss_format;
     double tenth;
     HRESULT ret = S_OK;
+    WAVEFORMATEXTENSIBLE *closest;
 
     tmp = oss_format = get_oss_format(fmt);
     if(oss_format < 0)
@@ -504,24 +506,35 @@
     if(fmt->nChannels == 0)
         return AUDCLNT_E_UNSUPPORTED_FORMAT;
 
+    closest = clone_format(fmt);
+    if(!closest)
+        return E_OUTOFMEMORY;
+
     tmp = fmt->nSamplesPerSec;
     if(ioctl(fd, SNDCTL_DSP_SPEED, &tmp) < 0){
         WARN("SPEED failed: %d (%s)\n", errno, strerror(errno));
+        free(closest);
         return E_FAIL;
     }
     tenth = fmt->nSamplesPerSec * 0.1;
     if(tmp > fmt->nSamplesPerSec + tenth || tmp < fmt->nSamplesPerSec - tenth){
         ret = S_FALSE;
+        closest->Format.nSamplesPerSec = tmp;
     }
 
     tmp = fmt->nChannels;
     if(ioctl(fd, SNDCTL_DSP_CHANNELS, &tmp) < 0){
         WARN("CHANNELS failed: %d (%s)\n", errno, strerror(errno));
+        free(closest);
         return E_FAIL;
     }
     if(tmp != fmt->nChannels){
         ret = S_FALSE;
-    }
+        closest->Format.nChannels = tmp;
+    }
+
+    if(closest->Format.wFormatTag == WAVE_FORMAT_EXTENSIBLE)
+        closest->dwChannelMask = get_channel_mask(closest->Format.nChannels);
 
     if(fmt->nBlockAlign != fmt->nChannels * fmt->wBitsPerSample / 8 ||
             fmt->nAvgBytesPerSec != fmt->nBlockAlign * fmt->nSamplesPerSec ||
@@ -534,6 +547,20 @@
         if(fmtex->dwChannelMask == 0 || fmtex->dwChannelMask & SPEAKER_RESERVED)
             ret = S_FALSE;
     }
+
+    if(ret == S_FALSE && !out)
+        ret = AUDCLNT_E_UNSUPPORTED_FORMAT;
+
+    if(ret == S_FALSE){
+        closest->Format.nBlockAlign =
+            closest->Format.nChannels * closest->Format.wBitsPerSample / 8;
+        closest->Format.nAvgBytesPerSec =
+            closest->Format.nBlockAlign * closest->Format.nSamplesPerSec;
+        if(closest->Format.wFormatTag == WAVE_FORMAT_EXTENSIBLE)
+            closest->Samples.wValidBitsPerSample = closest->Format.wBitsPerSample;
+        memcpy(out, closest, closest->Format.cbSize + sizeof(WAVEFORMATEX));
+    }
+    free(closest);
 
     TRACE("returning: %08x\n", (unsigned)ret);
     return ret;
@@ -585,7 +612,7 @@
     TRACE("min_channels: %d\n", ai.min_channels);
     TRACE("max_channels: %d\n", ai.max_channels);
 
-    params->result = setup_oss_device(params->share, stream->fd, params->fmt);
+    params->result = setup_oss_device(params->share, stream->fd, params->fmt, NULL);
     if(FAILED(params->result))
         goto exit;
 
@@ -1114,13 +1141,25 @@
     struct is_format_supported_params *params = args;
     int fd;
 
+    params->result = S_OK;
+
+    if(!params->fmt_in || (params->share == AUDCLNT_SHAREMODE_SHARED && !params->fmt_out))
+        params->result = E_POINTER;
+    else if(params->share != AUDCLNT_SHAREMODE_SHARED && params->share != AUDCLNT_SHAREMODE_EXCLUSIVE)
+        params->result = E_INVALIDARG;
+    else if(params->fmt_in->wFormatTag == WAVE_FORMAT_EXTENSIBLE &&
+            params->fmt_in->cbSize < sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX))
+        params->result = E_INVALIDARG;
+    if(FAILED(params->result))
+        return STATUS_SUCCESS;
+
     fd = open_device(params->device, params->flow);
     if(fd < 0){
         WARN("Unable to open device %s: %d (%s)\n", params->device, errno, strerror(errno));
         params->result = AUDCLNT_E_DEVICE_INVALIDATED;
         return STATUS_SUCCESS;
     }
-    params->result = setup_oss_device(params->share, fd, params->fmt_in);
+    params->result = setup_oss_device(params->share, fd, params->fmt_in, params->fmt_out);
     close(fd);
 
     return STATUS_SUCCESS;
@@ -1668,7 +1707,6 @@
     oss_is_started,
     oss_get_prop_value,
     oss_not_implemented,
-    oss_not_implemented,
     oss_midi_release,
     oss_midi_out_message,
     oss_midi_in_message,
@@ -1846,6 +1884,7 @@
         EDataFlow flow;
         AUDCLNT_SHAREMODE share;
         PTR32 fmt_in;
+        PTR32 fmt_out;
         HRESULT result;
     } *params32 = args;
     struct is_format_supported_params params =
@@ -1854,6 +1893,7 @@
         .flow = params32->flow,
         .share = params32->share,
         .fmt_in = ULongToPtr(params32->fmt_in),
+        .fmt_out = ULongToPtr(params32->fmt_out)
     };
     oss_is_format_supported(&params);
     params32->result = params.result;
@@ -2164,7 +2204,6 @@
     oss_is_started,
     oss_wow64_get_prop_value,
     oss_not_implemented,
-    oss_not_implemented,
     oss_midi_release,
     oss_wow64_midi_out_message,
     oss_wow64_midi_in_message,
