diff --git a/dlls/mfsrcsnk/media_source.c b/dlls/mfsrcsnk/media_source.c
--- a/dlls/mfsrcsnk/media_source.c
+++ b/dlls/mfsrcsnk/media_source.c
@@ -265,6 +265,7 @@
     LONG refcount;
 
     CRITICAL_SECTION cs;
+    IMFAsyncResult *shutdown_result;
     IMFMediaEventQueue *queue;
     IMFByteStream *stream;
     WCHAR *url;
@@ -274,6 +275,7 @@
     struct winedmo_demuxer winedmo_demuxer;
     struct winedmo_stream winedmo_stream;
     UINT64 file_size;
+    UINT64 position;
     INT64 duration;
     UINT stream_count;
     WCHAR mime_type[256];
@@ -493,7 +495,7 @@
     for (i = 0; i < source->stream_count; i++)
     {
         struct media_stream *stream = source->streams[i];
-        media_stream_start(stream, i, position);
+        media_stream_start(stream, source->stream_map[i], position);
     }
 
     if (position->vt == VT_I8 && (status = winedmo_demuxer_seek(source->winedmo_demuxer, position->hVal.QuadPart)))
@@ -684,7 +686,6 @@
     }
 
     queue_media_event_value(source->queue, MEEndOfPresentation, &empty);
-    source->state = SOURCE_STOPPED;
     return S_OK;
 }
 
@@ -1260,6 +1261,8 @@
         free(source->stream_map);
         free(source->streams);
 
+        if (source->shutdown_result)
+            IMFAsyncResult_Release(source->shutdown_result);
         IMFMediaEventQueue_Release(source->queue);
         IMFByteStream_Release(source->stream);
         free(source->url);
@@ -1387,9 +1390,6 @@
     TRACE("source %p, descriptor %p, format %s, position %s\n", source, descriptor,
             debugstr_guid(format), debugstr_propvar(position));
 
-    if (!format)
-        format = &GUID_NULL;
-
     EnterCriticalSection(&source->cs);
 
     if (source->state == SOURCE_SHUTDOWN)
@@ -1472,6 +1472,13 @@
         IMFMediaEventQueue_QueueEventParamVar(stream->queue, MEError, &GUID_NULL, MF_E_SHUTDOWN, NULL);
         IMFMediaEventQueue_Shutdown(stream->queue);
         IMFMediaStream_Release(&stream->IMFMediaStream_iface);
+    }
+
+    if (source->shutdown_result)
+    {
+        MFPutWorkItemEx(MFASYNC_CALLBACK_QUEUE_STANDARD, source->shutdown_result);
+        IMFAsyncResult_Release(source->shutdown_result);
+        source->shutdown_result = NULL;
     }
 
     LeaveCriticalSection(&source->cs);
@@ -1518,8 +1525,44 @@
 
     if (IsEqualGUID( &major, &MFMediaType_Video ))
         return media_type_from_mf_video_format( &format->video, media_type );
+
     if (IsEqualGUID( &major, &MFMediaType_Audio ))
-        return MFCreateAudioMediaType( &format->audio, (IMFAudioMediaType **)media_type );
+    {
+        const char *sgi = getenv("SteamGameId");
+        WAVEFORMATEXTENSIBLE *audio = (WAVEFORMATEXTENSIBLE *)&format->audio;
+
+        /* Warhammer 40,000: Dakka Squadron depends on the input format belonging to a specific set of formats.
+         * Append transcoded audio info to the user data so it can be restored, and create a fake AAC media
+         * type instead. If decoding support is added, PCM will work without a hack. */
+        if (sgi && !strcmp(sgi, "1253190") && format->audio.wFormatTag == WAVE_FORMAT_EXTENSIBLE
+                && IsEqualGUID(&audio->SubFormat, &MFAudioFormat_Vorbis))
+        {
+            size_t config_data_size = format->audio.cbSize + sizeof(WAVEFORMATEX) - sizeof(WAVEFORMATEXTENSIBLE);
+            size_t data_size = config_data_size + sizeof(WAVEFORMATEXTENSIBLE);
+            HEAACWAVEFORMAT *hwf;
+            HRESULT hr;
+
+            if (!(hwf = malloc(offsetof(HEAACWAVEFORMAT, pbAudioSpecificConfig[data_size]))))
+                return E_OUTOFMEMORY;
+
+            hwf->wfInfo.wfx = audio->Format;
+            hwf->wfInfo.wfx.wFormatTag = WAVE_FORMAT_MPEG_HEAAC;
+            hwf->wfInfo.wfx.cbSize = sizeof(HEAACWAVEINFO) + data_size - sizeof(WAVEFORMATEX);
+            hwf->wfInfo.wPayloadType = 0;
+            hwf->wfInfo.wAudioProfileLevelIndication = 0;
+            hwf->wfInfo.wStructType = 0;
+            hwf->wfInfo.wReserved1 = 0;
+            hwf->wfInfo.dwReserved2 = 0;
+            memcpy(hwf->pbAudioSpecificConfig, (BYTE *)(audio + 1), config_data_size);
+            memcpy(&hwf->pbAudioSpecificConfig[config_data_size], audio, sizeof(*audio));
+
+            hr = MFCreateAudioMediaType((WAVEFORMATEX *)hwf, (IMFAudioMediaType **)media_type);
+            free(hwf);
+            return hr;
+        }
+
+        return MFCreateAudioMediaType(&format->audio, (IMFAudioMediaType **)media_type);
+    }
 
     FIXME( "Unsupported major type %s\n", debugstr_guid( &major ) );
     return E_NOTIMPL;
@@ -1732,16 +1775,26 @@
 
     if (FAILED(IMFByteStream_Seek(source->stream, msoBegin, *pos, 0, pos)))
         return STATUS_UNSUCCESSFUL;
+
+    source->position = *pos;
     return STATUS_SUCCESS;
 }
 
 static NTSTATUS CDECL media_source_read_cb(struct winedmo_stream *stream, BYTE *buffer, ULONG *size)
 {
     struct media_source *source = CONTAINING_RECORD(stream, struct media_source, winedmo_stream);
+    UINT64 position;
+
     TRACE("stream %p, buffer %p, size %p\n", stream, buffer, size);
+
+    if (SUCCEEDED(IMFByteStream_GetCurrentPosition(source->stream, &position)) && position != source->position
+            && FAILED(IMFByteStream_SetCurrentPosition(source->stream, source->position)))
+        WARN("Failed to set current position\n");
 
     if (FAILED(IMFByteStream_Read(source->stream, buffer, *size, size)))
         return STATUS_UNSUCCESSFUL;
+
+    source->position += *size;
     return STATUS_SUCCESS;
 }
 
@@ -2035,7 +2088,7 @@
                        RRF_RT_REG_DWORD, NULL, &result, &size ))
         return !result;
 
-    return TRUE;
+    return FALSE;
 }
 
 static HRESULT WINAPI asf_byte_stream_plugin_factory_CreateInstance(IClassFactory *iface,
