diff --git a/dlls/mshtml/htmltable.c b/dlls/mshtml/htmltable.c
--- a/dlls/mshtml/htmltable.c
+++ b/dlls/mshtml/htmltable.c
@@ -24,7 +24,6 @@
 #include "winbase.h"
 #include "winuser.h"
 #include "ole2.h"
-#include "mshtmdid.h"
 
 #include "wine/debug.h"
 
@@ -431,22 +430,17 @@
     .get_attr_col          = HTMLElement_get_attr_col,
 };
 
-static void HTMLTableCellElement_init_dispex_info(dispex_data_t *info, compat_mode_t mode)
-{
-    static const dispex_hook_t hooks[] = {
-        {DISPID_IHTMLTABLECELL_CELLINDEX, .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    dispex_info_add_interface(info, IHTMLTableCell_tid, hooks);
-
-    HTMLElement_init_dispex_info(info, mode);
-}
+static const tid_t HTMLTableDataCellElement_iface_tids[] = {
+    IHTMLTableCell_tid,
+    0
+};
 
 dispex_static_data_t HTMLTableCellElement_dispex = {
-    .id           = OBJID_HTMLTableCellElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLTableCellElement,
+    .prototype_id = PROT_HTMLElement,
     .disp_tid     = DispHTMLTableCell_tid,
-    .init_info    = HTMLTableCellElement_init_dispex_info,
+    .iface_tids   = HTMLTableDataCellElement_iface_tids,
+    .init_info    = HTMLElement_init_dispex_info,
 };
 
 static const event_target_vtbl_t HTMLTableDataCellElement_event_target_vtbl = {
@@ -462,11 +456,12 @@
 };
 
 dispex_static_data_t HTMLTableDataCellElement_dispex = {
-    .id           = OBJID_HTMLTableDataCellElement,
-    .prototype_id = OBJID_HTMLTableCellElement,
+    .id           = PROT_HTMLTableDataCellElement,
+    .prototype_id = PROT_HTMLTableCellElement,
     .vtbl         = &HTMLTableDataCellElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLTableCell_tid,
-    .init_info    = HTMLTableCellElement_init_dispex_info,
+    .iface_tids   = HTMLTableDataCellElement_iface_tids,
+    .init_info    = HTMLElement_init_dispex_info,
 };
 
 HRESULT HTMLTableCell_Create(HTMLDocumentNode *doc, nsIDOMElement *nselem, HTMLElement **elem)
@@ -817,19 +812,6 @@
     .get_attr_col          = HTMLElement_get_attr_col,
 };
 
-static void HTMLTableRowElement_init_dispex_info(dispex_data_t *info, compat_mode_t mode)
-{
-    static const dispex_hook_t hooks[] = {
-        {DISPID_IHTMLTABLEROW_ROWINDEX,        .noattr = TRUE},
-        {DISPID_IHTMLTABLEROW_SECTIONROWINDEX, .noattr = TRUE},
-        {DISPID_IHTMLTABLEROW_CELLS,           .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    dispex_info_add_interface(info, IHTMLTableRow_tid, hooks);
-
-    HTMLElement_init_dispex_info(info, mode);
-}
-
 static const event_target_vtbl_t HTMLTableRowElement_event_target_vtbl = {
     {
         HTMLELEMENT_DISPEX_VTBL_ENTRIES,
@@ -842,12 +824,18 @@
     .handle_event       = HTMLElement_handle_event
 };
 
+static const tid_t HTMLTableRowElement_iface_tids[] = {
+    IHTMLTableRow_tid,
+    0
+};
+
 dispex_static_data_t HTMLTableRowElement_dispex = {
-    .id           = OBJID_HTMLTableRowElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLTableRowElement,
+    .prototype_id = PROT_HTMLElement,
     .vtbl         = &HTMLTableRowElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     =  DispHTMLTableRow_tid,
-    .init_info    = HTMLTableRowElement_init_dispex_info,
+    .iface_tids   = HTMLTableRowElement_iface_tids,
+    .init_info    = HTMLElement_init_dispex_info,
 };
 
 HRESULT HTMLTableRow_Create(HTMLDocumentNode *doc, nsIDOMElement *nselem, HTMLElement **elem)
@@ -1675,27 +1663,6 @@
     .get_attr_col          = HTMLElement_get_attr_col,
 };
 
-static void HTMLTableElement_init_dispex_info(dispex_data_t *info, compat_mode_t mode)
-{
-    static const dispex_hook_t table_hooks[] = {
-        {DISPID_IHTMLTABLE_ROWS,       .noattr = TRUE},
-        {DISPID_IHTMLTABLE_THEAD,      .noattr = TRUE},
-        {DISPID_IHTMLTABLE_TFOOT,      .noattr = TRUE},
-        {DISPID_IHTMLTABLE_TBODIES,    .noattr = TRUE},
-        {DISPID_IHTMLTABLE_CAPTION,    .noattr = TRUE},
-        {DISPID_IHTMLTABLE_READYSTATE, .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    static const dispex_hook_t table2_hooks[] = {
-        {DISPID_IHTMLTABLE2_CELLS,     .noattr = TRUE},
-        {DISPID_UNKNOWN}
-    };
-    dispex_info_add_interface(info, IHTMLTable_tid, table_hooks);
-    dispex_info_add_interface(info, IHTMLTable2_tid, table2_hooks);
-
-    HTMLElement_init_dispex_info(info, mode);
-}
-
 static const event_target_vtbl_t HTMLTableElement_event_target_vtbl = {
     {
         HTMLELEMENT_DISPEX_VTBL_ENTRIES,
@@ -1709,17 +1676,19 @@
 };
 
 static const tid_t HTMLTableElement_iface_tids[] = {
+    IHTMLTable_tid,
+    IHTMLTable2_tid,
     IHTMLTable3_tid,
     0
 };
 
 dispex_static_data_t HTMLTableElement_dispex = {
-    .id           = OBJID_HTMLTableElement,
-    .prototype_id = OBJID_HTMLElement,
+    .id           = PROT_HTMLTableElement,
+    .prototype_id = PROT_HTMLElement,
     .vtbl         = &HTMLTableElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLTable_tid,
     .iface_tids   = HTMLTableElement_iface_tids,
-    .init_info    = HTMLTableElement_init_dispex_info,
+    .init_info    = HTMLElement_init_dispex_info,
 };
 
 HRESULT HTMLTable_Create(HTMLDocumentNode *doc, nsIDOMElement *nselem, HTMLElement **elem)
