diff --git a/dlls/user32/tests/input.c b/dlls/user32/tests/input.c
--- a/dlls/user32/tests/input.c
+++ b/dlls/user32/tests/input.c
@@ -430,7 +430,6 @@
 static UINT (WINAPI *pGetRawInputDeviceInfoA) (HANDLE, UINT, void *, UINT *);
 static BOOL (WINAPI *pIsWow64Process)(HANDLE, PBOOL);
 static HKL (WINAPI *pLoadKeyboardLayoutEx)(HKL, const WCHAR *, UINT);
-static INT (WINAPI *pScheduleDispatchNotification)(HWND);
 
 /**********************adapted from input.c **********************************/
 
@@ -439,8 +438,6 @@
 static void init_function_pointers(void)
 {
     HMODULE hdll = GetModuleHandleA("user32");
-
-    pScheduleDispatchNotification = (void *)GetProcAddress(hdll, (LPCSTR)2582);
 
 #define GET_PROC(func) \
     if (!(p ## func = (void*)GetProcAddress(hdll, #func))) \
@@ -481,7 +478,9 @@
     ok_(file, line)( ret, "CreateProcessA failed, error %lu\n", GetLastError() );
     if (!ret) return;
 
-    wait_child_process( &info );
+    wait_child_process( info.hProcess );
+    CloseHandle( info.hThread );
+    CloseHandle( info.hProcess );
 }
 
 #define run_in_desktop( a, b, c ) run_in_desktop_( __FILE__, __LINE__, a, b, c )
@@ -511,7 +510,9 @@
     ok_(file, line)( ret, "CreateProcessA failed, error %lu\n", GetLastError() );
     if (!ret) return;
 
-    wait_child_process( &info );
+    wait_child_process( info.hProcess );
+    CloseHandle( info.hThread );
+    CloseHandle( info.hProcess );
 
     if (input)
     {
@@ -561,7 +562,7 @@
 }
 
 #define create_foreground_window( a ) create_foreground_window_( __FILE__, __LINE__, a, 5 )
-static HWND create_foreground_window_( const char *file, int line, BOOL fullscreen, UINT retries )
+HWND create_foreground_window_( const char *file, int line, BOOL fullscreen, UINT retries )
 {
     for (;;)
     {
@@ -3158,7 +3159,9 @@
     }
 
     SetEvent(process_ready);
-    winetest_wait_child_process(&process_info);
+    winetest_wait_child_process(process_info.hProcess);
+    CloseHandle(process_info.hProcess);
+    CloseHandle(process_info.hThread);
     CloseHandle(process_done);
     CloseHandle(process_start);
     CloseHandle(process_ready);
@@ -4310,8 +4313,8 @@
 
     mouse_event( MOUSEEVENTF_LEFTDOWN, 0, 0, 0, 0 );
     wait_messages( 5, FALSE );
-    button_down_hwnd[1].message.hwnd = hwnd;
-    ok_seq( button_down_hwnd );
+    button_down_hwnd_todo[1].message.hwnd = hwnd;
+    ok_seq( button_down_hwnd_todo );
     mouse_event( MOUSEEVENTF_LEFTUP, 0, 0, 0, 0 );
     wait_messages( 5, FALSE );
     button_up_hwnd[1].message.hwnd = hwnd;
@@ -6138,26 +6141,20 @@
     /* Test that the high word of the keyboard layout in CJK locale is the same as the low word,
      * even when IME is on */
     lang_id = PRIMARYLANGID(GetUserDefaultLCID());
-    switch (lang_id)
-    {
-    case LANG_CHINESE:
-        layout_name = "00000804";
-        break;
-    case LANG_JAPANESE:
-        layout_name = "00000411";
-        break;
-    case LANG_KOREAN:
-        layout_name = "00000412";
-        break;
-    default:
-        return;
-    }
-
-    hkl = GetKeyboardLayout(0);
-    ok(HIWORD(hkl) == LOWORD(hkl), "Got unexpected hkl %p.\n", hkl);
-
-    hkl = LoadKeyboardLayoutA(layout_name, 0);
-    ok(HIWORD(hkl) == LOWORD(hkl), "Got unexpected hkl %p.\n", hkl);
+    if (lang_id == LANG_CHINESE || lang_id == LANG_JAPANESE || lang_id == LANG_KOREAN)
+    {
+        hkl = GetKeyboardLayout(0);
+        ok(HIWORD(hkl) == LOWORD(hkl), "Got unexpected hkl %p.\n", hkl);
+
+        if (lang_id == LANG_CHINESE)
+            layout_name = "00000804";
+        else if (lang_id == LANG_JAPANESE)
+            layout_name = "00000411";
+        else if (lang_id == LANG_KOREAN)
+            layout_name = "00000412";
+        hkl = LoadKeyboardLayoutA(layout_name, 0);
+        ok(HIWORD(hkl) == LOWORD(hkl), "Got unexpected hkl %p.\n", hkl);
+    }
 }
 
 static void test_system_messages_with_rawinput_nolegacy(void)
@@ -6309,29 +6306,6 @@
 
     SetForegroundWindow( hwnd );
     DestroyWindow( hwnd );
-}
-
-static void test_ScheduleDispatchNotification(void)
-{
-    HWND hwnd;
-    INT ret;
-
-    if (!pScheduleDispatchNotification)
-    {
-        win_skip("ScheduleDispatchNotification is unavailable.\n");
-        return;
-    }
-
-    hwnd = CreateWindowW(L"static", NULL, WS_POPUP | WS_VISIBLE, 100, 100, 200, 200, NULL, NULL,
-                         NULL, NULL);
-
-    ret = pScheduleDispatchNotification(NULL);
-    ok(!ret, "Got unexpected %d.\n", ret);
-
-    ret = pScheduleDispatchNotification(hwnd);
-    ok(ret == 2, "Got unexpected %d.\n", ret);
-
-    DestroyWindow(hwnd);
 }
 
 START_TEST(input)
@@ -6378,7 +6352,6 @@
     test_OemKeyScan();
     test_rawinput(argv[0]);
     test_DefRawInputProc();
-    test_ScheduleDispatchNotification();
 
     if(pGetMouseMovePointsEx)
         test_GetMouseMovePointsEx( argv );
