diff --git a/dlls/user32/mdi.c b/dlls/user32/mdi.c
--- a/dlls/user32/mdi.c
+++ b/dlls/user32/mdi.c
@@ -306,12 +306,12 @@
 
             ci->hWindowMenu = hmenuWindow;
 
+            /* Add items to the new Window menu */
             ci->nActiveChildren = nActiveChildren_old;
+            MDI_RefreshMenu(ci);
         }
         else
             ci->hWindowMenu = hmenuWindow;
-
-        MDI_RefreshMenu(ci);
     }
 
     if (hmenuFrame)
@@ -686,7 +686,7 @@
     }
     HeapFree( GetProcessHeap(), 0, win_array );
 
-    if (has_icons) NtUserArrangeIconicWindows( client );
+    if (has_icons) ArrangeIconicWindows( client );
     return 0;
 }
 
@@ -773,7 +773,7 @@
         }
     }
     HeapFree( GetProcessHeap(), 0, win_array );
-    if (has_icons) NtUserArrangeIconicWindows( client );
+    if (has_icons) ArrangeIconicWindows( client );
 }
 
 /* ----------------------- Frame window ---------------------------- */
@@ -856,7 +856,7 @@
     NtUserSetMenuDefaultItem(hSysPopup, SC_CLOSE, FALSE);
 
     /* redraw menu */
-    NtUserDrawMenuBar(frame);
+    DrawMenuBar(frame);
 
     return TRUE;
 }
@@ -910,7 +910,7 @@
     /* minimize */
     NtUserDeleteMenu( menu, SC_MINIMIZE, MF_BYCOMMAND );
 
-    NtUserDrawMenuBar(frame);
+    DrawMenuBar(frame);
 
     return TRUE;
 }
@@ -1090,7 +1090,7 @@
 
       case WM_MDIICONARRANGE:
 	ci->mdiFlags |= MDIF_NEEDUPDATE;
-        NtUserArrangeIconicWindows( hwnd );
+        ArrangeIconicWindows( hwnd );
 	ci->sbRecalc = SB_BOTH+1;
         SendMessageW( hwnd, WM_MDICALCCHILDSCROLL, 0, 0 );
         return 0;
@@ -1357,7 +1357,6 @@
 	DefWindowProcA(hwnd, message, wParam, lParam);
 	if( ci->hwndChildMaximized == hwnd )
 	    MDI_UpdateFrameText( GetParent(client), client, TRUE, NULL );
-        MDI_RefreshMenu( ci );
         return 1; /* success. FIXME: check text length */
 
     case WM_GETMINMAXINFO:
@@ -1398,7 +1397,6 @@
         DefWindowProcW(hwnd, message, wParam, lParam);
         if( ci->hwndChildMaximized == hwnd )
             MDI_UpdateFrameText( GetParent(client), client, TRUE, NULL );
-        MDI_RefreshMenu( ci );
         return 1; /* success. FIXME: check text length */
 
     case WM_GETMINMAXINFO:
@@ -1660,7 +1658,7 @@
             {
                 HeapFree( GetProcessHeap(), 0, list );
                 NtUserShowScrollBar( hwnd, SB_BOTH, FALSE );
-                goto done;
+                return;
             }
             if (style & WS_VISIBLE)
             {
@@ -1702,8 +1700,6 @@
                         }
 			break;
     }
-
-done:
     SetThreadDpiAwarenessContext( context );
 }
 
@@ -1789,6 +1785,56 @@
                               SW_INVALIDATE | SW_ERASE | SW_SCROLLCHILDREN );
 done:
     SetThreadDpiAwarenessContext( context );
+}
+
+
+/******************************************************************************
+ *		CascadeWindows (USER32.@) Cascades MDI child windows
+ *
+ * RETURNS
+ *    Success: Number of cascaded windows.
+ *    Failure: 0
+ */
+WORD WINAPI
+CascadeWindows (HWND hwndParent, UINT wFlags, const RECT *lpRect,
+		UINT cKids, const HWND *lpKids)
+{
+    FIXME("(%p,0x%08x,...,%u,...): stub\n", hwndParent, wFlags, cKids);
+    return 0;
+}
+
+
+/***********************************************************************
+ *		CascadeChildWindows (USER32.@)
+ */
+WORD WINAPI CascadeChildWindows( HWND parent, UINT flags )
+{
+    return CascadeWindows( parent, flags, NULL, 0, NULL );
+}
+
+
+/******************************************************************************
+ *		TileWindows (USER32.@) Tiles MDI child windows
+ *
+ * RETURNS
+ *    Success: Number of tiled windows.
+ *    Failure: 0
+ */
+WORD WINAPI
+TileWindows (HWND hwndParent, UINT wFlags, const RECT *lpRect,
+	     UINT cKids, const HWND *lpKids)
+{
+    FIXME("(%p,0x%08x,...,%u,...): stub\n", hwndParent, wFlags, cKids);
+    return 0;
+}
+
+
+/***********************************************************************
+ *		TileChildWindows (USER32.@)
+ */
+WORD WINAPI TileChildWindows( HWND parent, UINT flags )
+{
+    return TileWindows( parent, flags, NULL, 0, NULL );
 }
 
 
