diff --git a/dlls/comdlg32/tests/itemdlg.c b/dlls/comdlg32/tests/itemdlg.c
--- a/dlls/comdlg32/tests/itemdlg.c
+++ b/dlls/comdlg32/tests/itemdlg.c
@@ -40,21 +40,6 @@
 static HRESULT (WINAPI *pSHCreateShellItem)(LPCITEMIDLIST,IShellFolder*,LPCITEMIDLIST,IShellItem**);
 static HRESULT (WINAPI *pSHGetIDListFromObject)(IUnknown*, PIDLIST_ABSOLUTE*);
 static HRESULT (WINAPI *pSHCreateItemFromParsingName)(PCWSTR,IBindCtx*,REFIID,void**);
-
-#define check_interface(a, b, c) check_interface_(__LINE__, a, b, c)
-static void check_interface_(unsigned int line, void *iface_ptr, REFIID iid, BOOL supported)
-{
-    IUnknown *iface = iface_ptr;
-    HRESULT hr, expected_hr;
-    IUnknown *unk;
-
-    expected_hr = supported ? S_OK : E_NOINTERFACE;
-
-    hr = IUnknown_QueryInterface(iface, iid, (void **)&unk);
-    ok_(__FILE__, line)(hr == expected_hr, "Got hr %#lx, expected %#lx.\n", hr, expected_hr);
-    if (SUCCEEDED(hr))
-        IUnknown_Release(unk);
-}
 
 static void init_function_pointers(void)
 {
@@ -349,6 +334,7 @@
 
 static BOOL test_instantiation(void)
 {
+    IFileDialog *pfd;
     IFileOpenDialog *pfod;
     IFileSaveDialog *pfsd;
     IServiceProvider *psp;
@@ -367,14 +353,9 @@
     }
     ok(hr == S_OK, "got 0x%08lx.\n", hr);
 
-    check_interface(pfod, &IID_IModalWindow, TRUE);
-    check_interface(pfod, &IID_IFileDialog, TRUE);
-    check_interface(pfod, &IID_IExplorerBrowserEvents, TRUE);
-    check_interface(pfod, &IID_ICommDlgBrowser3, TRUE);
-    check_interface(pfod, &IID_IFileSaveDialog, FALSE);
-    check_interface(pfod, &IID_IFileDialogEvents, FALSE);
-    check_interface(pfod, &IID_IExplorerBrowser, FALSE);
-    check_interface(pfod, &IID_IShellBrowser, FALSE);
+    hr = IFileOpenDialog_QueryInterface(pfod, &IID_IFileDialog, (void**)&pfd);
+    ok(hr == S_OK, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IFileDialog_Release(pfd);
 
     hr = IFileOpenDialog_QueryInterface(pfod, &IID_IFileDialogCustomize, (void**)&punk);
     ok(hr == S_OK, "got 0x%08lx.\n", hr);
@@ -384,6 +365,10 @@
     ok(punk == unk2, "got %p, %p\n", punk, unk2);
     IUnknown_Release(punk);
     IUnknown_Release(unk2);
+
+    hr = IFileOpenDialog_QueryInterface(pfod, &IID_IFileSaveDialog, (void**)&pfsd);
+    ok(hr == E_NOINTERFACE, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IFileSaveDialog_Release(pfsd);
 
     hr = IFileOpenDialog_QueryInterface(pfod, &IID_IServiceProvider, (void**)&psp);
     ok(hr == S_OK, "got 0x%08lx.\n", hr);
@@ -417,6 +402,26 @@
         IServiceProvider_Release(psp);
     }
 
+    hr = IFileOpenDialog_QueryInterface(pfod, &IID_IFileDialogEvents, (void**)&punk);
+    ok(hr == E_NOINTERFACE, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
+
+    hr = IFileOpenDialog_QueryInterface(pfod, &IID_IExplorerBrowser, (void**)&punk);
+    ok(hr == E_NOINTERFACE, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
+
+    hr = IFileOpenDialog_QueryInterface(pfod, &IID_IExplorerBrowserEvents, (void**)&punk);
+    ok(hr == S_OK, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
+
+    hr = IFileOpenDialog_QueryInterface(pfod, &IID_ICommDlgBrowser3, (void**)&punk);
+    ok(hr == S_OK, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
+
+    hr = IFileOpenDialog_QueryInterface(pfod, &IID_IShellBrowser, (void**)&punk);
+    ok(hr == E_NOINTERFACE, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
+
     hr = IFileOpenDialog_QueryInterface(pfod, &IID_IOleWindow, (void**)&pow);
     ok(hr == S_OK, "got 0x%08lx.\n", hr);
     if(SUCCEEDED(hr))
@@ -455,14 +460,9 @@
     }
     ok(hr == S_OK, "got 0x%08lx.\n", hr);
 
-    check_interface(pfsd, &IID_IModalWindow, TRUE);
-    check_interface(pfsd, &IID_IFileDialog, TRUE);
-    check_interface(pfsd, &IID_IExplorerBrowserEvents, TRUE);
-    check_interface(pfsd, &IID_ICommDlgBrowser3, TRUE);
-    check_interface(pfsd, &IID_IFileOpenDialog, FALSE);
-    check_interface(pfsd, &IID_IFileDialogEvents, FALSE);
-    check_interface(pfsd, &IID_IExplorerBrowser, FALSE);
-    check_interface(pfsd, &IID_IShellBrowser, FALSE);
+    hr = IFileSaveDialog_QueryInterface(pfsd, &IID_IFileDialog, (void**)&pfd);
+    ok(hr == S_OK, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IFileDialog_Release(pfd);
 
     hr = IFileSaveDialog_QueryInterface(pfsd, &IID_IFileDialogCustomize, (void**)&punk);
     ok(hr == S_OK, "got 0x%08lx.\n", hr);
@@ -472,6 +472,30 @@
     ok(punk == unk2, "got %p, %p\n", punk, unk2);
     IUnknown_Release(punk);
     IUnknown_Release(unk2);
+
+    hr = IFileSaveDialog_QueryInterface(pfsd, &IID_IFileOpenDialog, (void**)&pfod);
+    ok(hr == E_NOINTERFACE, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IFileOpenDialog_Release(pfod);
+
+    hr = IFileSaveDialog_QueryInterface(pfsd, &IID_IFileDialogEvents, (void**)&punk);
+    ok(hr == E_NOINTERFACE, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IFileDialog_Release(pfd);
+
+    hr = IFileSaveDialog_QueryInterface(pfsd, &IID_IExplorerBrowser, (void**)&punk);
+    ok(hr == E_NOINTERFACE, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
+
+    hr = IFileSaveDialog_QueryInterface(pfsd, &IID_IExplorerBrowserEvents, (void**)&punk);
+    ok(hr == S_OK, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
+
+    hr = IFileSaveDialog_QueryInterface(pfsd, &IID_ICommDlgBrowser3, (void**)&punk);
+    ok(hr == S_OK, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
+
+    hr = IFileSaveDialog_QueryInterface(pfsd, &IID_IShellBrowser, (void**)&punk);
+    ok(hr == E_NOINTERFACE, "got 0x%08lx.\n", hr);
+    if(SUCCEEDED(hr)) IUnknown_Release(punk);
 
     hr = IFileSaveDialog_QueryInterface(pfsd, &IID_IOleWindow, (void**)&pow);
     ok(hr == S_OK, "got 0x%08lx.\n", hr);
@@ -1141,11 +1165,15 @@
     const WCHAR filetype1_broken[] = {'f','n','a','m','e','1',' ', '(','*','.','t','x','t',')',0};
 
     cb_filetype = find_window(dlg_hwnd, NULL, filetype1);
+    ok(cb_filetype != NULL || broken(cb_filetype == NULL), "Could not find combobox on first attempt\n");
+
     if(!cb_filetype)
     {
-        /* Verified on Windows 10: the string "(*.txt)" is required to find the combobox. */
+        /* Not sure when this happens. Some specific version?
+         * Seen on 32-bit English Vista */
+        trace("Didn't find combobox on first attempt, trying broken string..\n");
         cb_filetype = find_window(dlg_hwnd, NULL, filetype1_broken);
-        ok(cb_filetype != NULL, "Failed to find combobox.\n");
+        ok(broken(cb_filetype != NULL), "Failed to find combobox on second attempt\n");
         if(!cb_filetype)
             return;
     }
