diff --git a/dlls/ntoskrnl.exe/tests/driver.c b/dlls/ntoskrnl.exe/tests/driver.c
--- a/dlls/ntoskrnl.exe/tests/driver.c
+++ b/dlls/ntoskrnl.exe/tests/driver.c
@@ -58,9 +58,6 @@
 static DRIVER_OBJECT *driver_obj;
 static DEVICE_OBJECT *lower_device, *upper_device;
 
-static IRP *queued_async_irps[2];
-static unsigned int queued_async_count;
-
 static POBJECT_TYPE *pExEventObjectType, *pIoFileObjectType, *pPsThreadType, *pIoDriverObjectType;
 static PEPROCESS *pPsInitialSystemProcess;
 static void *create_caller_thread;
@@ -287,22 +284,12 @@
 
     MmProbeAndLockPages(mdl, KernelMode, IoReadAccess);
 
-    addr = MmMapLockedPages(mdl, KernelMode);
-    ok(addr != NULL, "MmMapLockedPages failed\n");
-    if (addr != NULL)
-        ok(!kmemcmp(addr, buffer, sizeof(buffer)), "Unexpected data in mapped memory\n");
-
-    MmUnmapLockedPages(addr, mdl);
-
     addr = MmMapLockedPagesSpecifyCache(mdl, KernelMode, MmCached, NULL, FALSE, NormalPagePriority);
     todo_wine
     ok(addr != NULL, "MmMapLockedPagesSpecifyCache failed\n");
-    if (addr != NULL)
-        ok(!kmemcmp(addr, buffer, sizeof(buffer)), "Unexpected data in mapped memory\n");
 
     MmUnmapLockedPages(addr, mdl);
 
-    MmUnlockPages(mdl);
     IoFreeMdl(mdl);
 }
 
@@ -1081,14 +1068,12 @@
 }
 
 static int cancel_cnt;
-static HANDLE thread_id;
 
 static void WINAPI cancel_irp(DEVICE_OBJECT *device, IRP *irp)
 {
     IoReleaseCancelSpinLock(irp->CancelIrql);
     ok(irp->Cancel == TRUE, "Cancel = %x\n", irp->Cancel);
     ok(!irp->CancelRoutine, "CancelRoutine = %p\n", irp->CancelRoutine);
-    ok(PsGetCurrentThreadId() == thread_id, "Unexpected thread_id %p\n", PsGetCurrentThreadId());
     irp->IoStatus.Status = STATUS_CANCELLED;
     irp->IoStatus.Information = 0;
     cancel_cnt++;
@@ -1097,7 +1082,6 @@
 static void WINAPI cancel_ioctl_irp(DEVICE_OBJECT *device, IRP *irp)
 {
     IoReleaseCancelSpinLock(irp->CancelIrql);
-    todo_wine ok(PsGetCurrentThreadId() == thread_id, "Unexpected thread_id %p\n", PsGetCurrentThreadId());
     irp->IoStatus.Status = STATUS_CANCELLED;
     irp->IoStatus.Information = 0;
     cancel_cnt++;
@@ -1119,8 +1103,6 @@
     BOOL completion_called;
     BOOLEAN r;
     NTSTATUS status;
-
-    thread_id = PsGetCurrentThreadId();
 
     /* cancel IRP with no cancel routine */
     irp = IoBuildAsynchronousFsdRequest(IRP_MJ_FLUSH_BUFFERS, device, NULL, 0, NULL, &iosb);
@@ -2307,28 +2289,34 @@
     ok(!status, "got %#lx\n", status);
 
     attr.Attributes = 0;
-    status = ZwOpenDirectoryObject( &handle, DIRECTORY_ALL_ACCESS, &attr );
+    status = ZwOpenDirectoryObject( &handle, 0, &attr );
     ok(!status, "got %#lx\n", status);
     status = ZwMakeTemporaryObject( handle );
+    todo_wine
     ok(!status, "got %#lx\n", status);
     status = ZwMakeTemporaryObject( handle );
+    todo_wine
     ok(!status, "got %#lx\n", status);
     status = ZwClose( handle );
     ok(!status, "got %#lx\n", status);
     status = ZwOpenDirectoryObject( &handle, 0, &attr );
+    todo_wine
     ok(status == STATUS_OBJECT_NAME_NOT_FOUND, "got %#lx\n", status);
 
     status = ZwCreateDirectoryObject( &handle, GENERIC_ALL, &attr );
+    todo_wine
     ok(!status, "got %#lx\n", status);
     attr.Attributes = OBJ_PERMANENT;
-    status = ZwOpenDirectoryObject( &handle2, DIRECTORY_ALL_ACCESS, &attr );
+    status = ZwOpenDirectoryObject( &handle2, 0, &attr );
     ok(status == STATUS_SUCCESS, "got %#lx\n", status);
     status = ZwClose( handle2 );
     ok(!status, "got %#lx\n", status);
     status = ZwClose( handle );
+    todo_wine
     ok(!status, "got %#lx\n", status);
     attr.Attributes = 0;
     status = ZwOpenDirectoryObject( &handle, 0, &attr );
+    todo_wine
     ok(status == STATUS_OBJECT_NAME_NOT_FOUND, "got %#lx\n", status);
     if (!status) ZwClose( handle );
 }
@@ -2690,17 +2678,10 @@
     return STATUS_SUCCESS;
 }
 
-static void cancel_complete(IRP *irp)
-{
-    irp->IoStatus.Status = STATUS_SUCCESS;
-    IoCompleteRequest(irp, IO_NO_INCREMENT);
-}
-
 static NTSTATUS WINAPI driver_IoControl(DEVICE_OBJECT *device, IRP *irp)
 {
     IO_STACK_LOCATION *stack = IoGetCurrentIrpStackLocation(irp);
     NTSTATUS status = STATUS_NOT_SUPPORTED;
-    unsigned int i;
 
     switch (stack->Parameters.DeviceIoControl.IoControlCode)
     {
@@ -2714,36 +2695,13 @@
             status = test_load_driver_ioctl(irp, stack, &irp->IoStatus.Information);
             break;
         case IOCTL_WINETEST_RESET_CANCEL:
-            ok(!queued_async_count, "expected no queued asyncs left\n");
             cancel_cnt = 0;
             status = STATUS_SUCCESS;
             break;
         case IOCTL_WINETEST_TEST_CANCEL:
-            thread_id = PsGetCurrentThreadId();
             IoSetCancelRoutine(irp, cancel_ioctl_irp);
             IoMarkIrpPending(irp);
             return STATUS_PENDING;
-        case IOCTL_WINETEST_QUEUE_ASYNC:
-            if (queued_async_count >= ARRAY_SIZE(queued_async_irps))
-            {
-                ok(FALSE, "unexpected queued_async_count %u\n", queued_async_count);
-                status = ERROR_TOO_MANY_CMDS;
-                break;
-            }
-            queued_async_irps[queued_async_count++] = irp;
-            IoMarkIrpPending(irp);
-            return STATUS_PENDING;
-        case IOCTL_WINETEST_COMPLETE_ASYNC:
-            /* complete the pending IRPs gathered from
-             * IOCTL_WINETEST_QUEUE_ASYNC */
-            for (i = 0; i < queued_async_count; i++)
-            {
-                cancel_complete(queued_async_irps[i]);
-                queued_async_irps[i] = NULL;
-            }
-            queued_async_count = 0;
-            status = STATUS_SUCCESS;
-            break;
         case IOCTL_WINETEST_GET_CANCEL_COUNT:
             status = get_dword(irp, stack, &irp->IoStatus.Information, cancel_cnt);
             break;
