diff --git a/dlls/ntdll/unix/loadorder.c b/dlls/ntdll/unix/loadorder.c
--- a/dlls/ntdll/unix/loadorder.c
+++ b/dlls/ntdll/unix/loadorder.c
@@ -59,6 +59,7 @@
 static HANDLE app_key;
 static BOOL init_done;
 static BOOL main_exe_loaded;
+static BOOL eac_launcher_process;
 
 
 /***************************************************************************
@@ -362,11 +363,32 @@
  */
 void set_load_order_app_name( const WCHAR *app_name )
 {
+    static const WCHAR eac_launcherW[] = {'P','R','O','T','O','N','_','E','A','C','_','L','A','U','N','C','H','E','R','_','P','R','O','C','E','S','S',0};
+    static const WCHAR crossoutW[] = {'C','r','o','s','s','o','u','t','.','e','x','e',0};
     const WCHAR *p;
 
     if ((p = wcsrchr( app_name, '\\' ))) app_name = p + 1;
     app_key = open_app_key( app_name );
     main_exe_loaded = TRUE;
+
+    if (!wcscmp( app_name, crossoutW ))
+    {
+        ERR( "Disabling EAC bridge.\n" );
+        eac_launcher_process = TRUE;
+        return;
+    }
+
+    p = NtCurrentTeb()->Peb->ProcessParameters->Environment;
+    while(*p)
+    {
+        if (!wcsncmp( p, eac_launcherW, ARRAY_SIZE(eac_launcherW) - 1 ))
+        {
+            eac_launcher_process = TRUE;
+            break;
+        }
+
+        p += wcslen(p) + 1;
+    }
 }
 
 
@@ -378,37 +400,83 @@
  */
 enum loadorder get_load_order( const UNICODE_STRING *nt_name )
 {
+    static const WCHAR easyanticheat_x86W[] = {'e','a','s','y','a','n','t','i','c','h','e','a','t','_','x','8','6','.','d','l','l',0};
+    static const WCHAR easyanticheat_x64W[] = {'e','a','s','y','a','n','t','i','c','h','e','a','t','_','x','6','4','.','d','l','l',0};
+    static const WCHAR easyanticheatW[] = {'e','a','s','y','a','n','t','i','c','h','e','a','t','.','d','l','l',0};
+    static const WCHAR soW[] = {'s','o',0};
+
     static const WCHAR prefixW[] = {'\\','?','?','\\'};
     enum loadorder ret = LO_INVALID;
     const WCHAR *path = nt_name->Buffer;
-    unsigned int i, len = nt_name->Length / sizeof(WCHAR);
+    const WCHAR *p;
     WCHAR *module, *basename;
+    int len;
 
     if (!init_done) init_load_order();
 
-    if (len > 4 && !wcsncmp( path, prefixW, 4 ))
-    {
-        path += 4;
-        len -= 4;
+    if (!wcsncmp( path, prefixW, 4 )) path += 4;
+
+    TRACE("looking for %s\n", debugstr_w(path));
+
+    /* HACK: special logic for easyanticheat bridge: only load the bridge (builtin) if there exists a native version of the library next to the windows version */
+
+    basename = get_basename((WCHAR *)path);
+    if (!wcsicmp(basename, easyanticheat_x86W) || !wcsicmp(basename, easyanticheat_x64W) || !wcsicmp(basename, easyanticheatW))
+    {
+        UNICODE_STRING eac_unix_name, out_nt_name;
+        OBJECT_ATTRIBUTES attr;
+        char *unix_path = NULL;
+        NTSTATUS status;
+
+        if (eac_launcher_process)
+        {
+            ret = LO_NATIVE;
+            TRACE("got hardcoded %s for %s, as this is the EAC launcher process\n", debugstr_loadorder(ret), debugstr_w(path) );
+            return ret;
+        }
+
+        len = wcslen(nt_name->Buffer);
+        eac_unix_name.Buffer = malloc( (len + 5) * sizeof(WCHAR) );
+        wcscpy(eac_unix_name.Buffer, nt_name->Buffer);
+
+        basename = get_basename(eac_unix_name.Buffer);
+        if (!wcsicmp(basename, easyanticheatW))
+            wcscpy(basename, easyanticheat_x64W);
+        wcscpy(&basename[18], soW);
+        eac_unix_name.Length = eac_unix_name.MaximumLength = wcslen(eac_unix_name.Buffer) * sizeof(WCHAR);
+        InitializeObjectAttributes(&attr, &eac_unix_name, 0, NULL, NULL);
+
+        status = get_nt_and_unix_names( &attr, &out_nt_name, &unix_path, FILE_OPEN, FALSE );
+        free( out_nt_name.Buffer );
+        free( unix_path );
+        if (!status)
+        {
+            free(eac_unix_name.Buffer);
+            ret = LO_BUILTIN;
+            TRACE( "got hardcoded %s for %s, as the eac unix library is present\n", debugstr_loadorder(ret), debugstr_w(path) );
+            return ret;
+        }
+        else
+        {
+            ret = LO_NATIVE;
+            TRACE( "got hardcoded %s for %s, as the eac unix library (%s) is not present. status %x\n", debugstr_loadorder(ret), debugstr_w(path), debugstr_w(eac_unix_name.Buffer), (int)status );
+            free(eac_unix_name.Buffer);
+            return ret;
+        }
     }
 
     /* Strip path information if the module resides in the system directory
      */
-    if (len > wcslen(system_dir) - 4 && !wcsnicmp( system_dir + 4, path, wcslen(system_dir) - 4 ))
-    {
-        unsigned int pos = wcslen( system_dir ) - 4;
-        while (pos < len && (path[pos] == '\\' || path[pos] == '/')) pos++;
-        for (i = pos; i < len; i++) if (path[i] == '\\' || path[i] == '/') break;
-        if (i == len)
-        {
-            path += pos;
-            len -= pos;
-        }
-    }
-
+    if (!wcsnicmp( system_dir + 4, path, wcslen(system_dir) - 4 ))
+    {
+        p = path + wcslen( system_dir ) - 4;
+        while (*p == '\\' || *p == '/') p++;
+        if (!wcschr( p, '\\' ) && !wcschr( p, '/' )) path = p;
+    }
+
+    if (!(len = wcslen(path))) return ret;
     if (!(module = malloc( (len + 2) * sizeof(WCHAR) ))) return ret;
-    memcpy( module + 1, path, len * sizeof(WCHAR) );  /* reserve module[0] for the wildcard char */
-    module[len + 1] = 0;
+    wcscpy( module + 1, path );  /* reserve module[0] for the wildcard char */
     remove_dll_ext( module + 1 );
     basename = get_basename( module + 1 );
 
@@ -429,13 +497,13 @@
     if (!main_exe_loaded && basename != module+1)
     {
         ret = LO_NATIVE_BUILTIN;
-        TRACE( "got main exe default %s for %s\n", debugstr_loadorder(ret), debugstr_us(nt_name) );
+        TRACE( "got main exe default %s for %s\n", debugstr_loadorder(ret), debugstr_w(path) );
         goto done;
     }
 
     /* and last the hard-coded default */
     ret = LO_DEFAULT;
-    TRACE( "got hardcoded %s for %s\n", debugstr_loadorder(ret), debugstr_us(nt_name) );
+    TRACE( "got hardcoded %s for %s\n", debugstr_loadorder(ret), debugstr_w(path) );
 
  done:
     free( module );
