diff --git a/dlls/user32/tests/dialog.c b/dlls/user32/tests/dialog.c
--- a/dlls/user32/tests/dialog.c
+++ b/dlls/user32/tests/dialog.c
@@ -39,8 +39,6 @@
 #include "wingdi.h"
 #include "winuser.h"
 #include "winnls.h"
-
-static BOOL (WINAPI *pNtUserModifyUserStartupInfoFlags)(DWORD,DWORD);
 
 #define MAXHWNDS 1024
 static HWND hwnd [MAXHWNDS];
@@ -149,17 +147,6 @@
     {0, 0, 0, 0}
 };
 
-static void pump_messages(void)
-{
-    MSG msg;
-
-    while (PeekMessageA( &msg, 0, 0, 0, PM_REMOVE ))
-    {
-        TranslateMessage( &msg );
-        DispatchMessageA( &msg );
-    }
-}
-
 static DWORD get_button_style(HWND button)
 {
     return GetWindowLongW(button, GWL_STYLE) & BS_TYPEMASK;
@@ -1469,7 +1456,7 @@
 }
 
 static const WCHAR testtextW[] = {'W','n','d','T','e','x','t',0};
-static const char *testtext = "WndText\0\0";  /* add a WCHAR null since it can be passed directly to DefWindowProcW */
+static const char *testtext = "WndText";
 
 enum defdlgproc_text
 {
@@ -2261,63 +2248,7 @@
     return CallNextHookEx(NULL, code, wParam, lParam);
 }
 
-static LRESULT CALLBACK msgbox_hook_proc2(INT code, WPARAM wParam, LPARAM lParam)
-{
-    return 1; /* prevent message box or dialog window creation. */
-}
-
-static void test_message_box_startup_info(void)
-{
-    HHOOK hook;
-    HWND hwnd;
-    int ret;
-
-    hook = SetWindowsHookExA( WH_CBT, msgbox_hook_proc2, NULL, GetCurrentThreadId() );
-    ret = DialogBoxParamA( GetModuleHandleA( NULL ), "TEST_EMPTY_DIALOG", 0, NULL, 0 );
-    ok(ret == -1, "got %d\n", ret);
-    UnhookWindowsHookEx( hook );
-
-    hwnd = CreateWindowA( "static", "overlapped2", WS_OVERLAPPED, 0, 0, 0, 0, NULL, NULL, GetModuleHandleW(NULL), NULL );
-    ok( !!hwnd, "got NULL.\n" );
-    pump_messages();
-    ShowWindow( hwnd, SW_SHOWDEFAULT );
-    /* startup info window show flags are in effect. */
-    ret = IsWindowVisible( hwnd );
-    ok( !ret, "got %d.\n", ret );
-    DestroyWindow( hwnd );
-    pump_messages();
-
-    /* set startup info flags once again. */
-    pNtUserModifyUserStartupInfoFlags( STARTF_USESHOWWINDOW, STARTF_USESHOWWINDOW );
-    hook = SetWindowsHookExA( WH_CBT, msgbox_hook_proc2, NULL, GetCurrentThreadId() );
-    ret = MessageBoxA(NULL, "Text", "MSGBOX caption", msgbox_type);
-    todo_wine ok(!ret, "got %d\n", ret); /* Wine returns -1 here. */
-    UnhookWindowsHookEx( hook );
-
-    hwnd = CreateWindowA( "static", "overlapped2", WS_OVERLAPPED, 0, 0, 0, 0, NULL, NULL, GetModuleHandleW(NULL), NULL );
-    ok( !!hwnd, "got NULL.\n" );
-    pump_messages();
-    ShowWindow( hwnd, SW_SHOWDEFAULT );
-    ret = IsWindowVisible( hwnd );
-    /* startup info flags have no effect, while the message box window wasn't even created. */
-    ok( ret, "got %d.\n", ret );
-    DestroyWindow( hwnd );
-    pump_messages();
-
-    /* sanity check, set startup flags once againg and check that it works */
-    pNtUserModifyUserStartupInfoFlags( STARTF_USESHOWWINDOW, STARTF_USESHOWWINDOW );
-    hwnd = CreateWindowA( "static", "overlapped2", WS_OVERLAPPED, 0, 0, 0, 0, NULL, NULL, GetModuleHandleW(NULL), NULL );
-    ok( !!hwnd, "got NULL.\n" );
-    pump_messages();
-    ShowWindow( hwnd, SW_SHOWDEFAULT );
-    ret = IsWindowVisible( hwnd );
-    /* startup info flags have no effect, while the message box window wasn't even created. */
-    ok( !ret, "got %d.\n", ret );
-    DestroyWindow( hwnd );
-    pump_messages();
-}
-
-static void test_MessageBox(char **argv)
+static void test_MessageBox(void)
 {
     static const UINT tests[] =
     {
@@ -2329,9 +2260,6 @@
         MB_OKCANCEL | MB_TASKMODAL | MB_TOPMOST,
         MB_OKCANCEL | MB_TASKMODAL | MB_SYSTEMMODAL | MB_TOPMOST,
     };
-    STARTUPINFOA sa = {.cb = sizeof(STARTUPINFOA)};
-    PROCESS_INFORMATION info;
-    char cmdline[MAX_PATH];
     unsigned int i;
     HHOOK hook;
     int ret;
@@ -2350,18 +2278,6 @@
     }
 
     UnhookWindowsHookEx(hook);
-
-    if (!pNtUserModifyUserStartupInfoFlags)
-    {
-        win_skip("NtUserModifyUserStartupInfoFlags is not available.\n");
-        return;
-    }
-    sa.dwFlags = STARTF_USESHOWWINDOW;
-    sa.wShowWindow = SW_HIDE;
-    sprintf(cmdline, "%s %s message_box_startup_info", argv[0], argv[1]);
-    ret = CreateProcessA(NULL, cmdline, NULL, NULL, FALSE, 0, NULL, NULL, &sa, &info);
-    ok(ret, "got error %lu\n", GetLastError());
-    wait_child_process(&info);
 }
 
 static INT_PTR CALLBACK custom_test_dialog_proc(HWND hdlg, UINT msg, WPARAM wparam, LPARAM lparam)
@@ -2520,21 +2436,9 @@
 
 START_TEST(dialog)
 {
-    char **argv;
-    int argc = winetest_get_mainargs( &argv );
-    HMODULE win32u = GetModuleHandleA("win32u.dll");
-
-    pNtUserModifyUserStartupInfoFlags = (void*)GetProcAddress(win32u, "NtUserModifyUserStartupInfoFlags");
-
     g_hinst = GetModuleHandleA (0);
 
     if (!RegisterWindowClasses()) assert(0);
-
-    if (argc == 3 && !strcmp(argv[2], "message_box_startup_info"))
-    {
-        test_message_box_startup_info();
-        return;
-    }
 
     test_dialog_custom_data();
     test_GetNextDlgItem();
@@ -2549,6 +2453,6 @@
     test_MessageBoxFontTest();
     test_SaveRestoreFocus();
     test_timer_message();
-    test_MessageBox(argv);
+    test_MessageBox();
     test_capture_release();
 }
