diff --git a/dlls/win32u/ntuser_private.h b/dlls/win32u/ntuser_private.h
--- a/dlls/win32u/ntuser_private.h
+++ b/dlls/win32u/ntuser_private.h
@@ -38,20 +38,26 @@
     SYSTEM_TIMER_CARET = 0xffff,
 };
 
+struct user_object
+{
+    HANDLE       handle;
+    unsigned int type;
+};
+
 #define OBJ_OTHER_PROCESS ((void *)1)  /* returned by get_user_handle_ptr on unknown handles */
 
 typedef struct tagWND
 {
-    HWND               handle;        /* window full handle */
+    struct user_object obj;           /* object header */
     HWND               parent;        /* Window parent */
     HWND               owner;         /* Window owner */
     struct tagCLASS   *class;         /* Window class */
     struct dce        *dce;           /* DCE pointer */
     WNDPROC            winproc;       /* Window procedure */
+    UINT               tid;           /* Owner thread id */
     HINSTANCE          hInstance;     /* Window hInstance (from CreateWindow) */
     struct window_rects rects;        /* window rects in window DPI, relative to the parent client area */
     RECT               normal_rect;   /* Normal window rect saved when maximized/minimized */
-    RECT               present_rect;  /* present rect for exclusive fullscreen mode */
     POINT              min_pos;       /* Position for minimized window */
     POINT              max_pos;       /* Position for maximized window */
     WCHAR             *text;          /* Window text */
@@ -65,15 +71,13 @@
     HICON              hIcon;         /* window's icon */
     HICON              hIconSmall;    /* window's small icon */
     HICON              hIconSmall2;   /* window's secondary small icon, derived from hIcon */
-    BOOL               has_icons;     /* window's icons have been initialized */
     HIMC               imc;           /* window's input context */
+    UINT               dpi_context;   /* window DPI awareness context */
     struct window_surface *surface;   /* Window surface if any */
-    struct opengl_drawable *current_drawable; /* current GL client surface for this window */
-    struct opengl_drawable *unused_drawable; /* unused GL client surface for this window */
+    struct list        vulkan_surfaces; /* list of vulkan surfaces created for this window */
     struct tagDIALOGINFO *dlgInfo;    /* Dialog additional info (dialogs only) */
-    int                swap_interval; /* OpenGL surface swap interval */
     int                pixel_format;  /* Pixel format set by the graphics driver */
-    int                clip_clients;  /* Has client surfaces that needs to be clipped out */
+    int                internal_pixel_format; /* Internal pixel format set via WGL_WINE_pixel_format_passthrough */
     int                cbWndExtra;    /* class cbWndExtra at window creation */
     DWORD_PTR          userdata;      /* User private data */
     DWORD              wExtra[1];     /* Window extra bytes */
@@ -88,6 +92,8 @@
 #define WIN_NEEDS_SHOW_OWNEDPOPUP 0x0020 /* WM_SHOWWINDOW:SC_SHOW must be sent in the next ShowOwnedPopup call */
 #define WIN_CHILDREN_MOVED        0x0040 /* children may have moved, ignore stored positions */
 #define WIN_HAS_IME_WIN           0x0080 /* the window has been registered with imm32 */
+#define WIN_IS_ACTIVATING         0x0100 /* the window is being activated */
+#define WIN_IS_TOUCH              0x0200 /* the window has been registered for touch input */
 
 #define WND_OTHER_PROCESS ((WND *)1)  /* returned by get_win_ptr on unknown window handles */
 #define WND_DESKTOP       ((WND *)2)  /* returned by get_win_ptr on the desktop window */
@@ -97,6 +103,13 @@
 {
     return hwnd == HWND_BROADCAST || hwnd == HWND_TOPMOST;
 }
+
+struct touchinput_thread_data
+{
+    BYTE       index;            /* history index */
+    TOUCHINPUT current[8];       /* current touch state */
+    TOUCHINPUT history[128][8];  /* touches history buffer */
+};
 
 struct mouse_tracking_info
 {
@@ -113,7 +126,7 @@
 {
     struct ntuser_thread_info     client_info;            /* Data shared with client */
     HANDLE                        server_queue;           /* Handle to server-side queue */
-    HANDLE                        idle_event;             /* Handle to the process idle event */
+    DWORD                         last_getmsg_time;       /* Get/PeekMessage last request time */
     LONGLONG                      last_driver_time;       /* Get/PeekMessage driver event time */
     WORD                          hook_call_depth;        /* Number of recursively called hook procs */
     WORD                          hook_unicode;           /* Is current hook unicode? */
@@ -123,6 +136,7 @@
     HKL                           kbd_layout;             /* Current keyboard layout */
     UINT                          kbd_layout_id;          /* Current keyboard layout ID */
     struct hardware_msg_data     *rawinput;               /* Current rawinput message data */
+    struct touchinput_thread_data *touchinput;            /* touch input thread local buffer */
     UINT                          spy_indent;             /* Current spy indent */
     BOOL                          clipping_cursor;        /* thread is currently clipping */
     DWORD                         clipping_reset;         /* time when clipping was last reset */
@@ -189,13 +203,12 @@
 ULONG_PTR get_class_long_ptr( HWND hwnd, INT offset, BOOL ansi );
 WORD get_class_word( HWND hwnd, INT offset );
 DLGPROC get_dialog_proc( DLGPROC proc, BOOL ansi );
+ATOM get_int_atom_value( UNICODE_STRING *name );
 WNDPROC get_winproc( WNDPROC proc, BOOL ansi );
 void get_winproc_params( struct win_proc_params *params, BOOL fixup_ansi_dst );
 struct dce *get_class_dce( struct tagCLASS *class );
 struct dce *set_class_dce( struct tagCLASS *class, struct dce *dce );
-extern atom_t wine_server_add_atom( void *req, UNICODE_STRING *str );
-extern BOOL is_desktop_class( UNICODE_STRING *name );
-extern BOOL is_message_class( UNICODE_STRING *name );
+BOOL needs_ime_window( HWND hwnd );
 extern void register_builtin_classes(void);
 extern void register_desktop_class(void);
 
@@ -210,29 +223,40 @@
 HICON alloc_cursoricon_handle( BOOL is_icon );
 
 /* dce.c */
-extern void free_dce( struct dce *dce, HWND hwnd, struct list *drawables );
+extern void free_dce( struct dce *dce, HWND hwnd );
 extern void invalidate_dce( WND *win, const RECT *old_rect );
-extern BOOL is_cache_dc( HDC hdc );
 
 /* message.c */
-extern void check_for_events( UINT flags );
+struct peek_message_filter
+{
+    HWND hwnd;
+    UINT first;
+    UINT last;
+    UINT mask;
+    UINT flags;
+    BOOL internal;
+    BOOL waited;
+};
+
+extern int peek_message( MSG *msg, const struct peek_message_filter *filter );
 
 /* systray.c */
 extern LRESULT system_tray_call( HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam, void *data );
 
-/* opengl.c */
-extern BOOL set_dc_pixel_format_internal( HDC hdc, int format, struct list *drawables );
-extern void release_opengl_drawables( struct list *drawables );
-
 /* vulkan.c */
-extern struct vulkan_instance *vulkan_instance_create( const struct vulkan_instance_extensions *extensions );
+extern PFN_vkGetDeviceProcAddr p_vkGetDeviceProcAddr;
+extern PFN_vkGetInstanceProcAddr p_vkGetInstanceProcAddr;
+
+extern BOOL vulkan_init(void);
+extern void vulkan_update_surfaces( HWND hwnd );
+extern void vulkan_detach_surfaces( struct list *surfaces );
 
 /* window.c */
-HANDLE alloc_user_handle( void *ptr, unsigned short type );
-void *free_user_handle( HANDLE handle, unsigned short type );
-void *get_user_handle_ptr( HANDLE handle, unsigned short type );
+HANDLE alloc_user_handle( struct user_object *ptr, unsigned int type );
+void *free_user_handle( HANDLE handle, unsigned int type );
+void *get_user_handle_ptr( HANDLE handle, unsigned int type );
 void release_user_handle_ptr( void *ptr );
-void *next_thread_user_object( UINT tid, HANDLE *handle, unsigned short type );
+void *next_process_user_handle_ptr( HANDLE *handle, unsigned int type );
 UINT win_set_flags( HWND hwnd, UINT set_mask, UINT clear_mask );
 
 static inline UINT win_get_flags( HWND hwnd )
@@ -240,7 +264,6 @@
     return win_set_flags( hwnd, 0, 0 );
 }
 
-struct obj_locator get_window_class_locator( HWND hwnd );
 WND *get_win_ptr( HWND hwnd );
 BOOL is_child( HWND parent, HWND child );
 BOOL is_window( HWND hwnd );
