diff --git a/dlls/winhttp/tests/winhttp.c b/dlls/winhttp/tests/winhttp.c
--- a/dlls/winhttp/tests/winhttp.c
+++ b/dlls/winhttp/tests/winhttp.c
@@ -2202,7 +2202,7 @@
 "\r\n";
 
 static const char okmsg_length0[] =
-"HTTP/1.1  200  OK\r\n"
+"HTTP/1.1 200 OK\r\n"
 "Server: winetest\r\n"
 "Content-length: 0\r\n"
 "\r\n";
@@ -2285,16 +2285,10 @@
 "Upgrade: websocket\r\n"
 "Connection: Upgrade\r\n";
 
-static const char temp_redirectmsg[] =
+static const char redirectmsg[] =
 "HTTP/1.1 307 Temporary Redirect\r\n"
 "Content-Length: 0\r\n"
 "Location: /temporary\r\n"
-"Connection: close\r\n\r\n";
-
-static const char perm_redirectmsg[] =
-"HTTP/1.1 308 Permanent Redirect\r\n"
-"Content-Length: 0\r\n"
-"Location: /permanent\r\n"
 "Connection: close\r\n\r\n";
 
 static const char badreplyheadermsg[] =
@@ -2302,10 +2296,6 @@
 "Server: winetest\r\n"
 "SpaceAfterHdr  :   bad\r\n"
 "OkHdr: ok\r\n"
-"\r\n";
-
-static const char nostatustext[] =
-"HTTP/1.1 200\r\n"
 "\r\n";
 
 static const char proxy_pac[] =
@@ -2552,15 +2542,11 @@
             }
             else send(c, notokmsg, sizeof(notokmsg) - 1, 0);
         }
-        else if (strstr(buffer, "POST /redirect-temp"))
+        else if (strstr(buffer, "POST /redirect"))
         {
-            send(c, temp_redirectmsg, sizeof temp_redirectmsg - 1, 0);
+            send(c, redirectmsg, sizeof redirectmsg - 1, 0);
         }
-        else if (strstr(buffer, "POST /redirect-perm"))
-        {
-            send(c, perm_redirectmsg, sizeof perm_redirectmsg - 1, 0);
-        }
-        else if (strstr(buffer, "POST /temporary") || strstr(buffer, "POST /permanent"))
+        else if (strstr(buffer, "POST /temporary"))
         {
             char buf[32];
             recv(c, buf, sizeof(buf), 0);
@@ -2624,12 +2610,6 @@
             r = server_receive_request(c, buffer, sizeof(buffer));
             ok(!r, "got %d, buffer[0] %d.\n", r, buffer[0] );
         }
-        if (strstr(buffer, "GET /nostatustext"))
-        {
-            send(c, nostatustext, sizeof nostatustext - 1, 0);
-            r = server_receive_request(c, buffer, sizeof(buffer));
-            ok(!r, "got %d, buffer[0] %d.\n", r, buffer[0] );
-        }
         shutdown(c, 2);
         closesocket(c);
         c = -1;
@@ -3148,8 +3128,7 @@
 static void test_no_headers(int port)
 {
     HINTERNET ses, con, req;
-    DWORD error, size;
-    WCHAR buf[10];
+    DWORD error;
     BOOL ret;
 
     ses = WinHttpOpen(L"winetest", WINHTTP_ACCESS_TYPE_NO_PROXY, NULL, NULL, 0);
@@ -3177,23 +3156,6 @@
     }
 
     WinHttpCloseHandle(req);
-
-    req = WinHttpOpenRequest(con, NULL, L"/nostatustext", NULL, NULL, NULL, 0);
-    ok(req != NULL, "failed to open a request %lu\n", GetLastError());
-
-    ret = WinHttpSendRequest(req, NULL, 0, NULL, 0, 0, 0);
-    ok(ret, "got %lu\n", GetLastError());
-
-    ret = WinHttpReceiveResponse(req, NULL);
-    ok(ret, "got %lu\n", GetLastError());
-
-    memset(buf, 0xcc, sizeof(buf));
-    size = sizeof(buf);
-    ret = WinHttpQueryHeaders(req, WINHTTP_QUERY_STATUS_TEXT, NULL, buf, &size, NULL);
-    ok(ret, "got %lu\n", GetLastError());
-    ok(!buf[0], "got %x\n", buf[0]);
-
-    WinHttpCloseHandle(req);
     WinHttpCloseHandle(con);
     WinHttpCloseHandle(ses);
 }
@@ -3337,7 +3299,7 @@
     WinHttpCloseHandle(ses);
 }
 
-static void test_redirect(int port, const WCHAR *path, const WCHAR *target)
+static void test_redirect(int port)
 {
     HINTERNET ses, con, req;
     char buf[128];
@@ -3351,14 +3313,14 @@
     con = WinHttpConnect(ses, L"localhost", port, 0);
     ok(con != NULL, "failed to open a connection %lu\n", GetLastError());
 
-    req = WinHttpOpenRequest(con, L"POST", path, NULL, NULL, NULL, 0);
+    req = WinHttpOpenRequest(con, L"POST", L"/redirect", NULL, NULL, NULL, 0);
     ok(req != NULL, "failed to open a request %lu\n", GetLastError());
 
     url[0] = 0;
     size = sizeof(url);
     ret = WinHttpQueryOption(req, WINHTTP_OPTION_URL, url, &size);
     ok(ret, "got %lu\n", GetLastError());
-    swprintf(expected, ARRAY_SIZE(expected), L"http://localhost:%u%s", port, path);
+    swprintf(expected, ARRAY_SIZE(expected), L"http://localhost:%u/redirect", port);
     ok(!wcscmp(url, expected), "expected %s got %s\n", wine_dbgstr_w(expected), wine_dbgstr_w(url));
 
     ret = WinHttpSendRequest(req, NULL, 0, (void *)"data", sizeof("data"), sizeof("data"), 0);
@@ -3384,12 +3346,8 @@
     size = sizeof(url);
     ret = WinHttpQueryOption(req, WINHTTP_OPTION_URL, url, &size);
     ok(ret, "got %lu\n", GetLastError());
-    swprintf(expected, ARRAY_SIZE(expected), L"http://localhost:%u/%s", port, target);
-    ok(!wcscmp(url, expected) ||
-       broken(!!wcsstr(url, L"redirect-perm")), /* < Win10 */
-       "expected %s got %s\n", wine_dbgstr_w(expected), wine_dbgstr_w(url));
-    if (wcsstr(url, L"redirect-perm"))
-        goto cleanup;
+    swprintf(expected, ARRAY_SIZE(expected), L"http://localhost:%u/temporary", port);
+    ok(!wcscmp(url, expected), "expected %s got %s\n", wine_dbgstr_w(expected), wine_dbgstr_w(url));
 
     status = 0xdeadbeef;
     size = sizeof(status);
@@ -3409,7 +3367,6 @@
     ok(ret, "failed to read data %lu\n", GetLastError());
     ok(count == 128, "got %lu\n", count);
 
-cleanup:
     WinHttpCloseHandle(req);
     WinHttpCloseHandle(con);
     WinHttpCloseHandle(ses);
@@ -3631,7 +3588,7 @@
     ret = WinHttpSetOption(session, WINHTTP_OPTION_WEB_SOCKET_SEND_BUFFER_SIZE, &value, sizeof(DWORD));
     ok(ret, "got %lu\n", GetLastError());
 
-    request = WinHttpOpenRequest(connection, L"GET", L"/", NULL, NULL, NULL, WINHTTP_FLAG_SECURE);
+    request = WinHttpOpenRequest(connection, L"GET", L"/", NULL, NULL, NULL, 0);
     ok(request != NULL, "got %lu\n", GetLastError());
 
     size = 0xdeadbeef;
@@ -6343,8 +6300,7 @@
     test_request_path_escapes(si.port);
     test_passport_auth(si.port);
     test_websocket(si.port);
-    test_redirect(si.port, L"/redirect-temp", L"temporary");
-    test_redirect(si.port, L"/redirect-perm", L"permanent");
+    test_redirect(si.port);
     test_WinHttpGetProxyForUrl(si.port);
     test_connection_cache(si.port);
 
