diff --git a/dlls/mshtml/mshtml_private_iface.idl b/dlls/mshtml/mshtml_private_iface.idl
--- a/dlls/mshtml/mshtml_private_iface.idl
+++ b/dlls/mshtml/mshtml_private_iface.idl
@@ -125,7 +125,7 @@
     [propget, id(1)]
     HRESULT subtle([retval, out] IDispatch **subtle);
     [id(2)]
-    HRESULT getRandomValues([in] IDispatch *typedArray, [retval, out] IDispatch **ret);
+    HRESULT getRandomValues([in] VARIANT *typedArray, [retval, out] IDispatch **ret);
 }
 
 [
@@ -261,145 +261,6 @@
     HRESULT classList([retval, out] IDispatch **class_list);
 }
 
-const long DISPID_IWINEHTMLFORMPRIVATE_NOVALIDATE = 51;
-const long DISPID_IWINEHTMLFORMPRIVATE_CHECKVALIDITY = 52;
-[
-    odl,
-    oleautomation,
-    dual,
-    hidden,
-    uuid(8a11c39f-12e3-dcc3-14ec-479ba0bddac1)
-]
-interface IWineHTMLFormPrivate : IDispatch
-{
-    [propput, id(50)]
-    HRESULT enctype([in] BSTR v);
-    [propget, id(50)]
-    HRESULT enctype([retval, out] BSTR *ret);
-    [propput, id(DISPID_IWINEHTMLFORMPRIVATE_NOVALIDATE)]
-    HRESULT noValidate([in] VARIANT_BOOL v);
-    [propget, id(DISPID_IWINEHTMLFORMPRIVATE_NOVALIDATE)]
-    HRESULT noValidate([retval, out] VARIANT_BOOL *ret);
-    [id(DISPID_IWINEHTMLFORMPRIVATE_CHECKVALIDITY)]
-    HRESULT checkValidity([retval, out] VARIANT_BOOL *ret);
-}
-
-[
-    odl,
-    oleautomation,
-    dual,
-    hidden,
-    uuid(f356f6d2-f210-4f54-24c7-3f88fa3c20fc)
-]
-interface IWineHTMLInputPrivate : IDispatch
-{
-    [propput, id(50)]
-    HRESULT autofocus([in] VARIANT_BOOL v);
-    [propget, id(50)]
-    HRESULT autofocus([retval, out] VARIANT_BOOL *ret);
-    [propget, id(51)]
-    HRESULT validationMessage([retval, out] BSTR *ret);
-    [propget, id(52)]
-    HRESULT validity([retval, out] IDispatch **ret);
-    [propget, id(53)]
-    HRESULT willValidate([retval, out] VARIANT_BOOL *ret);
-    [id(54)]
-    HRESULT setCustomValidity([in] VARIANT *message);
-    [id(55)]
-    HRESULT checkValidity([retval, out] VARIANT_BOOL *ret);
-    [propput, id(56)]
-    HRESULT formAction([in] BSTR v);
-    [propget, id(56)]
-    HRESULT formAction([retval, out] BSTR *ret);
-    [propput, id(57)]
-    HRESULT formEnctype([in] BSTR v);
-    [propget, id(57)]
-    HRESULT formEnctype([retval, out] BSTR *ret);
-    [propput, id(58)]
-    HRESULT formMethod([in] BSTR v);
-    [propget, id(58)]
-    HRESULT formMethod([retval, out] BSTR *ret);
-    [propput, id(59)]
-    HRESULT formNoValidate([in] VARIANT_BOOL v);
-    [propget, id(59)]
-    HRESULT formNoValidate([retval, out] VARIANT_BOOL *ret);
-    [propput, id(60)]
-    HRESULT formTarget([in] BSTR v);
-    [propget, id(60)]
-    HRESULT formTarget([retval, out] BSTR *ret);
-}
-
-[
-    odl,
-    oleautomation,
-    dual,
-    hidden,
-    uuid(d26d41ef-79ef-f921-a8a2-65739537ff07)
-]
-interface IWineHTMLSelectPrivate : IDispatch
-{
-    [propput, id(50)]
-    HRESULT autofocus([in] VARIANT_BOOL v);
-    [propget, id(50)]
-    HRESULT autofocus([retval, out] VARIANT_BOOL *ret);
-    [propget, id(51)]
-    HRESULT validationMessage([retval, out] BSTR *ret);
-    [propget, id(52)]
-    HRESULT validity([retval, out] IDispatch **ret);
-    [propget, id(53)]
-    HRESULT willValidate([retval, out] VARIANT_BOOL *ret);
-    [id(54)]
-    HRESULT setCustomValidity([in] VARIANT *message);
-    [id(55)]
-    HRESULT checkValidity([retval, out] VARIANT_BOOL *ret);
-}
-
-[
-    odl,
-    oleautomation,
-    dual,
-    hidden,
-    uuid(27454a8f-bf7a-e47e-9ea8-2d7d1faad588)
-]
-interface IWineHTMLTextAreaPrivate : IDispatch
-{
-    [propput, id(50)]
-    HRESULT autofocus([in] VARIANT_BOOL v);
-    [propget, id(50)]
-    HRESULT autofocus([retval, out] VARIANT_BOOL *ret);
-    [propget, id(51)]
-    HRESULT validationMessage([retval, out] BSTR *ret);
-    [propget, id(52)]
-    HRESULT validity([retval, out] IDispatch **ret);
-    [propget, id(53)]
-    HRESULT willValidate([retval, out] VARIANT_BOOL *ret);
-    [id(54)]
-    HRESULT setCustomValidity([in] VARIANT *message);
-    [id(55)]
-    HRESULT checkValidity([retval, out] VARIANT_BOOL *ret);
-}
-
-[
-    odl,
-    oleautomation,
-    dual,
-    hidden,
-    uuid(307eb8f7-362b-6398-8c22-54c1c714efd8)
-]
-interface IWineHTMLObjectPrivate : IDispatch
-{
-    [propget, id(50)]
-    HRESULT validationMessage([retval, out] BSTR *ret);
-    [propget, id(51)]
-    HRESULT validity([retval, out] IDispatch **ret);
-    [propget, id(52)]
-    HRESULT willValidate([retval, out] VARIANT_BOOL *ret);
-    [id(53)]
-    HRESULT setCustomValidity([in] VARIANT *message);
-    [id(54)]
-    HRESULT checkValidity([retval, out] VARIANT_BOOL *ret);
-}
-
 [
     odl,
     oleautomation,
@@ -455,18 +316,6 @@
     HRESULT persisted([retval, out] VARIANT_BOOL *ret);
 }
 
-[
-    odl,
-    oleautomation,
-    dual,
-    hidden,
-    uuid(c985818a-0a28-3da3-4023-9cf70955d79d)
-]
-interface IWinePerformancePrivate : IDispatch
-{
-    [id(1)]
-    HRESULT now([retval, out] double *p);
-}
 
 const long DISPID_IWINEXMLHTTPREQUESTPRIVATE_RESPONSE = 1;
 const long DISPID_IWINEXMLHTTPREQUESTPRIVATE_RESPONSETYPE = 2;
