diff --git a/dlls/windowscodecs/fliprotate.c b/dlls/windowscodecs/fliprotate.c
--- a/dlls/windowscodecs/fliprotate.c
+++ b/dlls/windowscodecs/fliprotate.c
@@ -1,6 +1,5 @@
 /*
  * Copyright 2010 Vincent Povirk for CodeWeavers
- * Copyright 2024 Dmitry Timoshkov
  *
  * This library is free software; you can redistribute it and/or
  * modify it under the terms of the GNU Lesser General Public
@@ -38,7 +37,6 @@
     int flip_x;
     int flip_y;
     int swap_xy;
-    UINT bpp;
     CRITICAL_SECTION lock; /* must be held when initialized */
 } FlipRotator;
 
@@ -156,7 +154,8 @@
 {
     FlipRotator *This = impl_from_IWICBitmapFlipRotator(iface);
     HRESULT hr;
-    UINT y, srcy, width, height;
+    UINT y;
+    UINT srcy, srcwidth, srcheight;
     WICRect rc;
     WICRect rect;
 
@@ -164,11 +163,21 @@
 
     if (!This->source) return WINCODEC_ERR_WRONGSTATE;
 
-    hr = IWICBitmapFlipRotator_GetSize(iface, &width, &height);
+    if (This->swap_xy || This->flip_x)
+    {
+        /* This requires knowledge of the pixel format. */
+        FIXME("flipping x and rotating are not implemented\n");
+        return E_NOTIMPL;
+    }
+
+    hr = IWICBitmapSource_GetSize(This->source, &srcwidth, &srcheight);
     if (FAILED(hr)) return hr;
 
     if (!prc)
     {
+        UINT width, height;
+        hr = IWICBitmapFlipRotator_GetSize(iface, &width, &height);
+        if (FAILED(hr)) return hr;
         rect.X = 0;
         rect.Y = 0;
         rect.Width = width;
@@ -176,53 +185,10 @@
         prc = &rect;
     }
 
-    if (This->swap_xy || This->flip_x)
-    {
-        UINT bytes_per_pixel = This->bpp / 8;
-        UINT srcx, x;
-
-        if (This->bpp < 8)
-        {
-            FIXME("Flipping x and rotating are not implemented for %u bpp bitmap\n", This->bpp);
-            return E_NOTIMPL;
-        }
-
-        for (y = prc->Y; y - prc->Y < prc->Height; y++)
-        {
-            BYTE *dst = pbBuffer;
-
-            if (This->flip_y)
-                srcy = height - 1 - y;
-            else
-                srcy = y;
-
-            for (x = prc->X; x - prc->X < prc->Width; x++)
-            {
-                if (This->flip_x)
-                    srcx = width - 1 - x;
-                else
-                    srcx = x;
-
-                rc.X = This->swap_xy ? srcy : srcx;
-                rc.Y = This->swap_xy ? srcx : srcy;
-                rc.Width = 1;
-                rc.Height = 1;
-                hr = IWICBitmapSource_CopyPixels(This->source, &rc, bytes_per_pixel, bytes_per_pixel, dst);
-                if (FAILED(hr)) return hr;
-
-                dst += bytes_per_pixel;
-            }
-
-            pbBuffer += cbStride;
-        }
-
-        return hr;
-    }
-
     for (y=prc->Y; y - prc->Y < prc->Height; y++)
     {
         if (This->flip_y)
-            srcy = height - 1 - y;
+            srcy = srcheight - 1 - y;
         else
             srcy = y;
 
@@ -246,7 +212,6 @@
     IWICBitmapSource *pISource, WICBitmapTransformOptions options)
 {
     FlipRotator *This = impl_from_IWICBitmapFlipRotator(iface);
-    GUID pf;
     HRESULT hr=S_OK;
 
     TRACE("(%p,%p,%u)\n", iface, pISource, options);
@@ -259,46 +224,23 @@
         goto end;
     }
 
-    switch (options & 3)
-    {
-    case WICBitmapTransformRotate0:
-        break;
-
-    case WICBitmapTransformRotate90:
+    if (options&WICBitmapTransformRotate90)
+    {
         This->swap_xy = 1;
-        This->flip_x = 1;
-        break;
-
-    case WICBitmapTransformRotate180:
-        This->flip_x = 1;
-        This->flip_y = 1;
-        break;
-
-    case WICBitmapTransformRotate270:
-        This->swap_xy = 1;
-        This->flip_y = 1;
-        break;
+        This->flip_x = !This->flip_x;
+    }
+
+    if (options&WICBitmapTransformRotate180)
+    {
+        This->flip_x = !This->flip_x;
+        This->flip_y = !This->flip_y;
     }
 
     if (options&WICBitmapTransformFlipHorizontal)
-    {
-        if (This->swap_xy)
-            This->flip_y = !This->flip_y;
-        else
-            This->flip_x = !This->flip_x;
-    }
+        This->flip_x = !This->flip_x;
 
     if (options&WICBitmapTransformFlipVertical)
-    {
-        if (This->swap_xy)
-            This->flip_x = !This->flip_x;
-        else
-            This->flip_y = !This->flip_y;
-    }
-
-    hr = IWICBitmapSource_GetPixelFormat(pISource, &pf);
-    if (SUCCEEDED(hr))
-        hr = get_pixelformat_bpp(&pf, &This->bpp);
+        This->flip_y = !This->flip_y;
 
     IWICBitmapSource_AddRef(pISource);
     This->source = pISource;
@@ -334,7 +276,6 @@
     This->flip_x = 0;
     This->flip_y = 0;
     This->swap_xy = 0;
-    This->bpp = 0;
     InitializeCriticalSectionEx(&This->lock, 0, RTL_CRITICAL_SECTION_FLAG_FORCE_DEBUG_INFO);
     This->lock.DebugInfo->Spare[0] = (DWORD_PTR)(__FILE__ ": FlipRotator.lock");
 
