diff --git a/dlls/kernelbase/thread.c b/dlls/kernelbase/thread.c
--- a/dlls/kernelbase/thread.c
+++ b/dlls/kernelbase/thread.c
@@ -33,6 +33,7 @@
 #include "wine/exception.h"
 #include "wine/asm.h"
 #include "wine/debug.h"
+#include "wine/heap.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(thread);
 
@@ -63,22 +64,6 @@
     return CreateRemoteThreadEx( process, sa, stack, start, param, flags, NULL, id );
 }
 
-struct proc_thread_attr
-{
-    DWORD_PTR attr;
-    SIZE_T size;
-    void *value;
-};
-
-struct _PROC_THREAD_ATTRIBUTE_LIST
-{
-    DWORD mask;  /* bitmask of items in list */
-    DWORD size;  /* max number of items in list */
-    DWORD count; /* number of items in list */
-    DWORD pad;
-    DWORD_PTR unk;
-    struct proc_thread_attr attrs[];
-};
 
 /***************************************************************************
  *           CreateRemoteThreadEx   (kernelbase.@)
@@ -88,78 +73,33 @@
                                                       LPVOID param, DWORD flags,
                                                       LPPROC_THREAD_ATTRIBUTE_LIST attributes, DWORD *id )
 {
-    ULONG_PTR buffer[offsetof( PS_ATTRIBUTE_LIST, Attributes[3] ) / sizeof(ULONG_PTR)];
-    PS_ATTRIBUTE_LIST *attr_list = (PS_ATTRIBUTE_LIST *)buffer;
-    struct _ACTIVATION_CONTEXT *actctx;
     HANDLE handle;
     CLIENT_ID client_id;
-    TEB *teb;
-    ULONG count = 0, ret;
-    OBJECT_ATTRIBUTES attr;
     SIZE_T stack_reserve = 0, stack_commit = 0;
-    GROUP_AFFINITY *group_affinity = NULL;
-
-    if (attributes)
-    {
-        DWORD i;
-        for (i = 0; i < attributes->count; i++)
-            switch (attributes->attrs[i].attr)
-            {
-            case PROC_THREAD_ATTRIBUTE_GROUP_AFFINITY:
-                group_affinity = attributes->attrs[i].value;
-                break;
-            default:
-                FIXME("thread attributes %Ix ignored\n", attributes->attrs[i].attr);
-                break;
-            }
-    }
+
+    if (attributes) FIXME("thread attributes ignored\n");
 
     if (flags & STACK_SIZE_PARAM_IS_A_RESERVATION) stack_reserve = stack;
     else stack_commit = stack;
 
-    attr_list->Attributes[count].Attribute    = PS_ATTRIBUTE_CLIENT_ID;
-    attr_list->Attributes[count].Size         = sizeof(client_id);
-    attr_list->Attributes[count].ValuePtr     = &client_id;
-    attr_list->Attributes[count].ReturnLength = NULL;
-    count++;
-    attr_list->Attributes[count].Attribute    = PS_ATTRIBUTE_TEB_ADDRESS;
-    attr_list->Attributes[count].Size         = sizeof(teb);
-    attr_list->Attributes[count].ValuePtr     = &teb;
-    attr_list->Attributes[count].ReturnLength = NULL;
-    count++;
-    if (group_affinity)
-    {
-        attr_list->Attributes[count].Attribute    = PS_ATTRIBUTE_GROUP_AFFINITY;
-        attr_list->Attributes[count].Size         = sizeof(*group_affinity);
-        attr_list->Attributes[count].ValuePtr     = group_affinity;
-        attr_list->Attributes[count].ReturnLength = NULL;
-        count++;
-    }
-    attr_list->TotalLength = offsetof( PS_ATTRIBUTE_LIST, Attributes[count] );
-
-    InitializeObjectAttributes( &attr, NULL, 0, NULL, sa ? sa->lpSecurityDescriptor : NULL );
-    if (sa && sa->bInheritHandle) attr.Attributes |= OBJ_INHERIT;
-
-    RtlGetActiveActivationContext( &actctx );
-
-    if (!set_ntstatus( NtCreateThreadEx( &handle, THREAD_ALL_ACCESS, &attr, process,
-                                         (PRTL_THREAD_START_ROUTINE)start, param,
-                                         THREAD_CREATE_FLAGS_CREATE_SUSPENDED, 0,
-                                         stack_commit, stack_reserve, attr_list )))
-    {
-        if (actctx) RtlReleaseActivationContext( actctx );
+    if (!set_ntstatus( RtlCreateUserThread( process, sa ? sa->lpSecurityDescriptor : NULL, TRUE,
+                                            0, stack_reserve, stack_commit,
+                                            (PRTL_THREAD_START_ROUTINE)start, param, &handle, &client_id )))
         return 0;
-    }
-
-    if (actctx)
-    {
-        ULONG_PTR cookie;
-        RtlActivateActivationContextEx( 0, teb, actctx, &cookie );
-        RtlReleaseActivationContext( actctx );
-    }
 
     if (id) *id = HandleToULong( client_id.UniqueThread );
-    if (!(flags & CREATE_SUSPENDED)) NtResumeThread( handle, &ret );
+    if (sa && sa->nLength >= sizeof(*sa) && sa->bInheritHandle)
+        SetHandleInformation( handle, HANDLE_FLAG_INHERIT, HANDLE_FLAG_INHERIT );
+    if (!(flags & CREATE_SUSPENDED))
+    {
+        ULONG ret;
+        if (NtResumeThread( handle, &ret ))
+        {
+            NtClose( handle );
+            SetLastError( ERROR_NOT_ENOUGH_MEMORY );
+            handle = 0;
+        }
+    }
     return handle;
 }
 
@@ -334,7 +274,7 @@
     if (!set_ntstatus( NtQueryInformationThread( thread, ThreadBasicInformation,
                                                  &info, sizeof(info), NULL )))
         return THREAD_PRIORITY_ERROR_RETURN;
-    return info.BasePriority;
+    return info.Priority;
 }
 
 
@@ -426,15 +366,6 @@
 DWORD WINAPI DECLSPEC_HOTPATCH QueueUserAPC( PAPCFUNC func, HANDLE thread, ULONG_PTR data )
 {
     return set_ntstatus( NtQueueApcThread( thread, call_user_apc, (ULONG_PTR)func, data, 0 ));
-}
-
-
-/***********************************************************************
- *	     QueueUserAPC2   (kernelbase.@)
- */
-DWORD WINAPI DECLSPEC_HOTPATCH QueueUserAPC2( PAPCFUNC func, HANDLE thread, ULONG_PTR data, QUEUE_USER_APC_FLAGS flags )
-{
-    return set_ntstatus( NtQueueApcThreadEx2( thread, NULL, flags, call_user_apc, (ULONG_PTR)func, data, 0 ));
 }
 
 
@@ -510,7 +441,7 @@
     if (status != STATUS_BUFFER_TOO_SMALL)
         return HRESULT_FROM_NT(status);
 
-    if (!(info = HeapAlloc( GetProcessHeap(), 0, length )))
+    if (!(info = heap_alloc( length )))
         return HRESULT_FROM_NT(STATUS_NO_MEMORY);
 
     status = NtQueryInformationThread( thread, ThreadNameInformation, info, length, &length );
@@ -526,7 +457,7 @@
         }
     }
 
-    HeapFree(GetProcessHeap(), 0, info);
+    heap_free(info);
 
     return HRESULT_FROM_NT(status);
 }
@@ -690,10 +621,6 @@
 DWORD WINAPI DECLSPEC_HOTPATCH SuspendThread( HANDLE thread )
 {
     DWORD ret;
-
-    /* NT in Win9x mode returns 0 for current thread. */
-    if ((GetVersion() & 0x80000000) && GetThreadId( thread ) == GetCurrentThreadId())
-        return 0;
 
     if (!set_ntstatus( NtSuspendThread( thread, &ret ))) ret = ~0U;
     return ret;
