diff --git a/dlls/msi/tests/action.c b/dlls/msi/tests/action.c
--- a/dlls/msi/tests/action.c
+++ b/dlls/msi/tests/action.c
@@ -1800,13 +1800,6 @@
     "Feature\tFeature\n"
     "assembly\t\t\tassembly feature\t1\t2\tMSITESTDIR\t0\n";
 
-static const char pa_feature_with_2nd_feature_dat[] =
-    "Feature\tFeature_Parent\tTitle\tDescription\tDisplay\tLevel\tDirectory_\tAttributes\n"
-    "s38\tS38\tL64\tL255\tI2\ti2\tS72\ti2\n"
-    "Feature\tFeature\n"
-    "assembly\t\t\tassembly feature\t1\t2\tMSITESTDIR\t0\n"
-    "2nd_feature\t\tSecond feature\tSpace filler\t1\t2\tMSITESTDIR\t0\n";
-
 static const char pa_feature_comp_dat[] =
     "Feature_\tComponent_\n"
     "s38\ts72\n"
@@ -2343,21 +2336,6 @@
     ADD_TABLE(property)
 };
 
-static const msi_table pa_tables_with_2nd_feature[] =
-{
-    ADD_TABLE(directory),
-    ADD_TABLE(pa_component),
-    ADD_TABLE(pa_feature_with_2nd_feature),
-    ADD_TABLE(pa_feature_comp),
-    ADD_TABLE(pa_file),
-    ADD_TABLE(pa_msi_assembly),
-    ADD_TABLE(pa_msi_assembly_name),
-    ADD_TABLE(pa_install_exec_seq),
-    ADD_TABLE(pa_custom_action),
-    ADD_TABLE(media),
-    ADD_TABLE(property)
-};
-
 static const msi_table rep_tables[] =
 {
     ADD_TABLE(directory),
@@ -2490,7 +2468,7 @@
     ok_(__FILE__, line)(!res, "Failed to query value, error %ld\n", res);
     ok_(__FILE__, line)(type == REG_MULTI_SZ, "Got wrong type %lu\n", type);
     ok_(__FILE__, line)(size == expect_size, "expected size %lu, got %lu\n", expect_size, size);
-    ok_(__FILE__, line)(size == expect_size && !memcmp(val, expect, size), "got %s\n", debugstr_an(val, size));
+    ok_(__FILE__, line)(!memcmp(val, expect, size), "got %s\n", debugstr_an(val, size));
 }
 
 static void check_reg_dword(HKEY prodkey, LPCSTR name, DWORD expected, BOOL todo, DWORD line)
@@ -3453,8 +3431,8 @@
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
     CHECK_REG_STR(props, "ProductID", "none");
-    if (company) CHECK_REG_STR(props, "RegCompany", company);
-    if (owner) CHECK_REG_STR(props, "RegOwner", owner);
+    CHECK_REG_STR(props, "RegCompany", company);
+    CHECK_REG_STR(props, "RegOwner", owner);
 
     RegDeleteValueA(props, "ProductID");
     RegDeleteValueA(props, "RegCompany");
@@ -3475,8 +3453,8 @@
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
     CHECK_REG_STR(props, "ProductID", "none");
-    if (company) CHECK_REG_STR(props, "RegCompany", company);
-    if (owner) CHECK_REG_STR(props, "RegOwner", owner);
+    CHECK_REG_STR(props, "RegCompany", company);
+    CHECK_REG_STR(props, "RegOwner", owner);
 
     RegDeleteValueA(props, "ProductID");
     RegDeleteValueA(props, "RegCompany");
@@ -4785,10 +4763,10 @@
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
     res = RegSetValueExA(hkey, "Value6", 0, REG_MULTI_SZ, (const BYTE *)"one\0", 5);
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
-    res = reg_set_str(hkey, "Value7", "one");
+    res = RegSetValueExA(hkey, "Value7", 0, REG_SZ, (const BYTE *)"one", 4);
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
-    res = reg_set_str(hkey, "instremove", "val");
+    res = RegSetValueExA(hkey, "instremove", 0, REG_SZ, (const BYTE *)"val", 3);
     ok(!res, "got %ld\n", res);
 
     res = RegCreateKeyA(hkey, "instremove", &subkey);
@@ -4941,13 +4919,13 @@
                           0, NULL, 0, KEY_ALL_ACCESS, NULL, &env2, NULL);
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
-    res = reg_set_str(env, "MSITESTVAR1", "0");
+    res = RegSetValueExA(env, "MSITESTVAR1", 0, REG_SZ, (const BYTE *)"0", 2);
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
-    res = reg_set_str(env, "MSITESTVAR2", "0");
+    res = RegSetValueExA(env, "MSITESTVAR2", 0, REG_SZ, (const BYTE *)"0", 2);
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
-    res = reg_set_str(env, "MSITESTVAR21", "1");
+    res = RegSetValueExA(env, "MSITESTVAR21", 0, REG_SZ, (const BYTE *)"1", 2);
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
     MsiSetInternalUI(INSTALLUILEVEL_NONE, NULL);
@@ -4997,10 +4975,10 @@
     CHECK_REG_STR(env, "MSITESTVAR24", "TestService");
     CHECK_REG_STR(env2, "MSITESTVAR100", "1");
 
-    res = reg_set_str(env, "MSITESTVAR22", "1");
+    res = RegSetValueExA(env, "MSITESTVAR22", 0, REG_SZ, (const BYTE *)"1", 2);
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
-    res = reg_set_str(env, "MSITESTVAR23", "1");
+    res = RegSetValueExA(env, "MSITESTVAR23", 0, REG_SZ, (const BYTE *)"1", 2);
     ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
 
     res = RegDeleteValueA(env, "MSITESTVAR25");
@@ -6422,75 +6400,6 @@
     res = RegOpenKeyA(HKEY_CLASSES_ROOT, path, &hkey);
     ok(res == ERROR_FILE_NOT_FOUND, "Expected ERROR_FILE_NOT_FOUND, got %ld\n", res);
 
-    /* When there are more than 1 feature in MSI package, the feature has to be recorded in
-       the assembly publication entry. */
-    DeleteFileA(msifile);
-    create_database(msifile, pa_tables_with_2nd_feature, ARRAY_SIZE(pa_tables_with_2nd_feature));
-
-    r = MsiInstallProductA(msifile, "ALLUSERS=1");
-    ok(r == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %u\n", r);
-
-    access = KEY_QUERY_VALUE;
-    res = RegOpenKeyExA(HKEY_CLASSES_ROOT, classes_path_dotnet, 0, access, &hkey);
-    if (res == ERROR_FILE_NOT_FOUND && is_wow64) /* Vista WOW64 */
-    {
-        trace("Using 64-bit registry view for HKCR\\Installer\n");
-        access = KEY_QUERY_VALUE | KEY_WOW64_64KEY;
-        res = RegOpenKeyExA(HKEY_CLASSES_ROOT, classes_path_dotnet, 0, access, &hkey);
-    }
-    ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
-    CHECK_REG_MULTI(hkey, name_dotnet, "rcHQPHq?CA@Uv-XqMI1eassembly>Z'q,T*76M@=YEg6My?~]\0");
-    RegCloseKey(hkey);
-
-    path = (is_wow64 || is_64bit) ? classes_path_dotnet_local_wow64 : classes_path_dotnet_local;
-    res = RegOpenKeyExA(HKEY_CLASSES_ROOT, path, 0, access, &hkey);
-    ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
-    CHECK_REG_MULTI(hkey, name_dotnet_local, "rcHQPHq?CA@Uv-XqMI1eassembly>LF,8A?0d.AW@vcZ$Cgox\0");
-    RegCloseKey(hkey);
-
-    res = RegOpenKeyExA(HKEY_CLASSES_ROOT, classes_path_win32, 0, access, &hkey);
-    ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
-    CHECK_REG_MULTI(hkey, name_win32, "rcHQPHq?CA@Uv-XqMI1eassembly>}NJjwR'%D9v1p!v{WV(%\0");
-    RegCloseKey(hkey);
-
-    path = (is_wow64 || is_64bit) ? classes_path_win32_local_wow64 : classes_path_win32_local;
-    res = RegOpenKeyExA(HKEY_CLASSES_ROOT, path, 0, access, &hkey);
-    ok(res == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %ld\n", res);
-    CHECK_REG_MULTI(hkey, name_win32_local, "rcHQPHq?CA@Uv-XqMI1eassembly>C)Uvlj*53A)u(QQ9=)X!\0");
-    RegCloseKey(hkey);
-
-    /* No registration is done for a local assembly with no matching file */
-    path = (is_wow64 || is_64bit) ? classes_path_fake_local_wow64 : classes_path_fake_local;
-    res = RegOpenKeyExA(HKEY_CLASSES_ROOT, path, 0, access, &hkey);
-    ok(res == ERROR_FILE_NOT_FOUND, "got %ld\n", res);
-
-    r = MsiInstallProductA(msifile, "REMOVE=ALL ALLUSERS=1");
-    ok(r == ERROR_SUCCESS, "Expected ERROR_SUCCESS, got %u\n", r);
-
-    res = RegOpenKeyA(HKEY_CLASSES_ROOT, classes_path_dotnet, &hkey);
-    if (res == ERROR_SUCCESS)
-    {
-        res = RegDeleteValueA(hkey, name_dotnet);
-        ok(res == ERROR_FILE_NOT_FOUND, "Expected ERROR_FILE_NOT_FOUND, got %ld\n", res);
-        RegCloseKey(hkey);
-    }
-
-    path = (is_wow64 || is_64bit) ? classes_path_dotnet_local_wow64 : classes_path_dotnet_local;
-    res = RegOpenKeyA(HKEY_CLASSES_ROOT, path, &hkey);
-    ok(res == ERROR_FILE_NOT_FOUND, "Expected ERROR_FILE_NOT_FOUND, got %ld\n", res);
-
-    res = RegOpenKeyA(HKEY_CLASSES_ROOT, classes_path_win32, &hkey);
-    if (res == ERROR_SUCCESS)
-    {
-        res = RegDeleteValueA(hkey, name_win32);
-        ok(res == ERROR_FILE_NOT_FOUND, "Expected ERROR_FILE_NOT_FOUND, got %ld\n", res);
-        RegCloseKey(hkey);
-    }
-
-    path = (is_wow64 || is_64bit) ? classes_path_win32_local_wow64 : classes_path_win32_local;
-    res = RegOpenKeyA(HKEY_CLASSES_ROOT, path, &hkey);
-    ok(res == ERROR_FILE_NOT_FOUND, "Expected ERROR_FILE_NOT_FOUND, got %ld\n", res);
-
 done:
     DeleteFileA("msitest\\win32.txt");
     DeleteFileA("msitest\\win32_local.txt");
