diff --git a/dlls/winegstreamer/wma_decoder.c b/dlls/winegstreamer/wma_decoder.c
--- a/dlls/winegstreamer/wma_decoder.c
+++ b/dlls/winegstreamer/wma_decoder.c
@@ -30,7 +30,8 @@
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(wmadec);
-WINE_DECLARE_DEBUG_CHANNEL(winediag);
+
+extern const GUID MFAudioFormat_XMAudio2;
 
 static const GUID *const wma_decoder_input_types[] =
 {
@@ -38,6 +39,7 @@
     &MFAudioFormat_WMAudioV8,
     &MFAudioFormat_WMAudioV9,
     &MFAudioFormat_WMAudio_Lossless,
+    &MFAudioFormat_XMAudio2,
 };
 static const GUID *const wma_decoder_output_types[] =
 {
@@ -504,7 +506,18 @@
 
 static HRESULT WINAPI transform_ProcessMessage(IMFTransform *iface, MFT_MESSAGE_TYPE message, ULONG_PTR param)
 {
-    FIXME("iface %p, message %#x, param %p stub!\n", iface, message, (void *)param);
+    struct wma_decoder *decoder = impl_from_IMFTransform(iface);
+
+    TRACE("iface %p, message %#x, param %p.\n", iface, message, (void *)param);
+
+    if (!decoder->wg_transform)
+        return MF_E_TRANSFORM_TYPE_NOT_SET;
+
+    if (message == MFT_MESSAGE_COMMAND_DRAIN)
+        return wg_transform_drain(decoder->wg_transform);
+
+    FIXME("Ignoring message %#x.\n", message);
+
     return S_OK;
 }
 
@@ -552,7 +565,7 @@
         return MF_E_TRANSFORM_NEED_MORE_INPUT;
     }
 
-    if (SUCCEEDED(hr = wg_transform_read_mf(decoder->wg_transform, samples->pSample, &samples->dwStatus, NULL)))
+    if (SUCCEEDED(hr = wg_transform_read_mf(decoder->wg_transform, samples->pSample, 0, &samples->dwStatus, NULL)))
         wg_sample_queue_flush(decoder->wg_sample_queue, false);
 
     return hr;
@@ -704,7 +717,6 @@
         const DMO_MEDIA_TYPE *type, DWORD flags)
 {
     struct wma_decoder *decoder = impl_from_IMediaObject(iface);
-    unsigned int i;
 
     TRACE("iface %p, index %lu, type %p, flags %#lx.\n", iface, index, type, flags);
 
@@ -732,12 +744,6 @@
     if (!IsEqualGUID(&type->majortype, &MEDIATYPE_Audio))
         return DMO_E_TYPE_NOT_ACCEPTED;
 
-    for (i = 0; i < ARRAY_SIZE(wma_decoder_input_types); ++i)
-        if (IsEqualGUID(&type->subtype, wma_decoder_input_types[i]))
-            break;
-    if (i == ARRAY_SIZE(wma_decoder_input_types))
-        return DMO_E_TYPE_NOT_ACCEPTED;
-
     if (flags & DMO_SET_TYPEF_TEST_ONLY)
         return S_OK;
 
@@ -751,6 +757,22 @@
         decoder->wg_transform = 0;
     }
 
+{
+    const char *sgi = getenv("SteamGameId");
+    if (sgi && (0
+        || !strcmp(sgi, "802870")
+        || !strcmp(sgi, "1230140")
+        || !strcmp(sgi, "2515070")
+        || !strcmp(sgi, "3625380")
+    ))
+    {
+        WAVEFORMATEX *wfx = (WAVEFORMATEX *)decoder->input_type.pbFormat;
+        decoder->input_type.subtype = MFAudioFormat_PCM;
+        wfx->wFormatTag = WAVE_FORMAT_PCM;
+        wfx->wBitsPerSample = 16;
+    }
+}
+
     return S_OK;
 }
 
@@ -759,7 +781,6 @@
 {
     struct wma_decoder *decoder = impl_from_IMediaObject(iface);
     struct wg_transform_attrs attrs = {0};
-    wg_transform_t new_transform = 0;
     unsigned int i;
     HRESULT hr;
 
@@ -797,30 +818,22 @@
 
     if (IsEqualGUID(&decoder->input_type.majortype, &GUID_NULL))
         return DMO_E_TYPE_NOT_SET;
-
-    if (!IsEqualGUID(&type->formattype, &FORMAT_WaveFormatEx))
-        return DMO_E_TYPE_NOT_ACCEPTED;
-
-    if (((WAVEFORMATEX *)decoder->input_type.pbFormat)->nChannels !=
-            ((WAVEFORMATEX *)type->pbFormat)->nChannels)
-        return DMO_E_TYPE_NOT_ACCEPTED;
-
-    if (FAILED(hr = wg_transform_create_quartz(&decoder->input_type, type, &attrs, &new_transform)))
-        return hr;
-
     if (flags & DMO_SET_TYPEF_TEST_ONLY)
-    {
-        wg_transform_destroy(new_transform);
         return S_OK;
-    }
 
     MoFreeMediaType(&decoder->output_type);
     memset(&decoder->output_type, 0, sizeof(decoder->output_type));
     MoCopyMediaType(&decoder->output_type, type);
 
     /* Set up wg_transform. */
-    if (decoder->wg_transform) wg_transform_destroy(decoder->wg_transform);
-    decoder->wg_transform = new_transform;
+    if (decoder->wg_transform)
+    {
+        wg_transform_destroy(decoder->wg_transform);
+        decoder->wg_transform = 0;
+    }
+    if (FAILED(hr = wg_transform_create_quartz(&decoder->input_type, &decoder->output_type,
+            &attrs, &decoder->wg_transform)))
+        return hr;
 
     return S_OK;
 }
@@ -924,8 +937,8 @@
 
 static HRESULT WINAPI media_object_AllocateStreamingResources(IMediaObject *iface)
 {
-    TRACE("iface %p.\n", iface);
-    return S_OK;
+    FIXME("iface %p stub!\n", iface);
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI media_object_FreeStreamingResources(IMediaObject *iface)
@@ -973,10 +986,9 @@
         buffers[0].dwStatus |= DMO_OUTPUT_DATA_BUFFERF_INCOMPLETE;
         wg_sample_queue_flush(decoder->wg_sample_queue, false);
     }
-    else if (hr == MF_E_TRANSFORM_NEED_MORE_INPUT)
-        hr = S_FALSE;
-
-
+
+    if (hr == MF_E_TRANSFORM_NEED_MORE_INPUT)
+        return S_FALSE;
     return hr;
 }
 
@@ -1061,26 +1073,10 @@
 
 HRESULT wma_decoder_create(IUnknown *outer, IUnknown **out)
 {
-    static const WAVEFORMATEX output_format =
-    {
-        .wFormatTag = WAVE_FORMAT_IEEE_FLOAT, .wBitsPerSample = 32, .nSamplesPerSec = 44100, .nChannels = 1,
-    };
-    static const WMAUDIO2WAVEFORMAT input_format =
-    {
-        .wfx = {.wFormatTag = WAVE_FORMAT_WMAUDIO2, .wBitsPerSample = 16, .nSamplesPerSec = 44100, .nChannels = 1,
-                .nAvgBytesPerSec = 3000, .nBlockAlign = 139, .cbSize = sizeof(input_format) - sizeof(WAVEFORMATEX)},
-        .wEncodeOptions = 1,
-    };
     struct wma_decoder *decoder;
     HRESULT hr;
 
     TRACE("outer %p, out %p.\n", outer, out);
-
-    if (FAILED(hr = check_audio_transform_support(&input_format.wfx, &output_format)))
-    {
-        ERR_(winediag)("GStreamer doesn't support WMA decoding, please install appropriate plugins.\n");
-        return hr;
-    }
 
     if (!(decoder = calloc(1, sizeof(*decoder))))
         return E_OUTOFMEMORY;
