diff --git a/dlls/windows.ui/uisettings.c b/dlls/windows.ui/uisettings.c
--- a/dlls/windows.ui/uisettings.c
+++ b/dlls/windows.ui/uisettings.c
@@ -18,23 +18,17 @@
  */
 
 #include "private.h"
-#include "initguid.h"
-#include "weakref.h"
 
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(ui);
-
-static EventRegistrationToken dummy_cookie = {.value = 0xdeadbeef};
 
 struct uisettings
 {
     IUISettings IUISettings_iface;
     IUISettings2 IUISettings2_iface;
     IUISettings3 IUISettings3_iface;
-    IUISettings4 IUISettings4_iface;
-    IUISettings5 IUISettings5_iface;
-    struct weak_reference_source weak_reference_source;
+    LONG ref;
 };
 
 static inline struct uisettings *impl_from_IUISettings( IUISettings *iface )
@@ -65,18 +59,6 @@
     {
         *out = &impl->IUISettings3_iface;
     }
-    else if (IsEqualGUID( iid, &IID_IUISettings4 ))
-    {
-        *out = &impl->IUISettings4_iface;
-    }
-    else if (IsEqualGUID( iid, &IID_IUISettings5 ))
-    {
-        *out = &impl->IUISettings5_iface;
-    }
-    else if (IsEqualGUID( iid, &IID_IWeakReferenceSource ))
-    {
-        *out = &impl->weak_reference_source.IWeakReferenceSource_iface;
-    }
 
     if (!*out)
     {
@@ -91,7 +73,7 @@
 static ULONG WINAPI uisettings_AddRef( IUISettings *iface )
 {
     struct uisettings *impl = impl_from_IUISettings( iface );
-    ULONG ref = weak_reference_strong_add_ref( &impl->weak_reference_source );
+    ULONG ref = InterlockedIncrement( &impl->ref );
     TRACE( "iface %p, ref %lu.\n", iface, ref );
     return ref;
 }
@@ -99,7 +81,7 @@
 static ULONG WINAPI uisettings_Release( IUISettings *iface )
 {
     struct uisettings *impl = impl_from_IUISettings( iface );
-    ULONG ref = weak_reference_strong_release( &impl->weak_reference_source );
+    ULONG ref = InterlockedDecrement( &impl->ref );
 
     TRACE( "iface %p, ref %lu.\n", iface, ref );
 
@@ -375,14 +357,13 @@
 static HRESULT WINAPI uisettings3_add_ColorValuesChanged( IUISettings3 *iface, ITypedEventHandler_UISettings_IInspectable *handler, EventRegistrationToken *cookie )
 {
     FIXME( "iface %p, handler %p, cookie %p stub!\n", iface, handler, cookie );
-    *cookie = dummy_cookie;
-    return S_OK;
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI uisettings3_remove_ColorValuesChanged( IUISettings3 *iface, EventRegistrationToken cookie )
 {
     FIXME( "iface %p, cookie %#I64x stub!\n", iface, cookie.value );
-    return S_OK;
+    return E_NOTIMPL;
 }
 
 static const struct IUISettings3Vtbl uisettings3_vtbl =
@@ -400,81 +381,6 @@
     uisettings3_remove_ColorValuesChanged,
 };
 
-DEFINE_IINSPECTABLE( uisettings4, IUISettings4, struct uisettings, IUISettings_iface );
-
-static HRESULT WINAPI uisettings4_get_AdvancedEffectsEnabled( IUISettings4 *iface, boolean *value )
-{
-    FIXME( "iface %p, value %p stub!.\n", iface, value );
-    *value = TRUE;
-    return S_OK;
-}
-
-static HRESULT WINAPI uisettings4_add_AdvancedEffectsEnabledChanged( IUISettings4 *iface, ITypedEventHandler_UISettings_IInspectable *handler, EventRegistrationToken *cookie )
-{
-    FIXME( "iface %p, handler %p, cookie %p stub!\n", iface, handler, cookie );
-    *cookie = dummy_cookie;
-    return S_OK;
-}
-
-static HRESULT WINAPI uisettings4_remove_AdvancedEffectsEnabledChanged( IUISettings4 *iface, EventRegistrationToken cookie )
-{
-    FIXME( "iface %p, cookie %#I64x stub!\n", iface, cookie.value );
-    return S_OK;
-}
-
-static const struct IUISettings4Vtbl uisettings4_vtbl =
-{
-    uisettings4_QueryInterface,
-    uisettings4_AddRef,
-    uisettings4_Release,
-    /* IInspectable methods */
-    uisettings4_GetIids,
-    uisettings4_GetRuntimeClassName,
-    uisettings4_GetTrustLevel,
-    /* IUISettings4 methods */
-    uisettings4_get_AdvancedEffectsEnabled,
-    uisettings4_add_AdvancedEffectsEnabledChanged,
-    uisettings4_remove_AdvancedEffectsEnabledChanged,
-};
-
-DEFINE_IINSPECTABLE( uisettings5, IUISettings5, struct uisettings, IUISettings_iface );
-
-static HRESULT WINAPI uisettings5_get_AutoHideScrollBars( IUISettings5 *iface, boolean *value )
-{
-    FIXME( "iface %p, value %p stub!.\n", iface, value );
-    *value = FALSE;
-    return S_OK;
-}
-
-static HRESULT WINAPI uisettings5_add_AutoHideScrollBarsChanged( IUISettings5 *iface, ITypedEventHandler_UISettings_UISettingsAutoHideScrollBarsChangedEventArgs *handler,
-                                                                 EventRegistrationToken *cookie )
-{
-    FIXME( "iface %p, handler %p, cookie %p stub!\n", iface, handler, cookie );
-    *cookie = dummy_cookie;
-    return S_OK;
-}
-
-static HRESULT WINAPI uisettings5_remove_AutoHideScrollBarsChanged( IUISettings5 *iface, EventRegistrationToken cookie )
-{
-    FIXME( "iface %p, cookie %#I64x stub!\n", iface, cookie.value );
-    return S_OK;
-}
-
-static const struct IUISettings5Vtbl uisettings5_vtbl =
-{
-    uisettings5_QueryInterface,
-    uisettings5_AddRef,
-    uisettings5_Release,
-    /* IInspectable methods */
-    uisettings5_GetIids,
-    uisettings5_GetRuntimeClassName,
-    uisettings5_GetTrustLevel,
-    /* IUISettings5 methods */
-    uisettings5_get_AutoHideScrollBars,
-    uisettings5_add_AutoHideScrollBarsChanged,
-    uisettings5_remove_AutoHideScrollBarsChanged,
-};
-
 struct uisettings_statics
 {
     IActivationFactory IActivationFactory_iface;
@@ -543,7 +449,6 @@
 static HRESULT WINAPI factory_ActivateInstance( IActivationFactory *iface, IInspectable **instance )
 {
     struct uisettings *impl;
-    HRESULT hr;
 
     TRACE( "iface %p, instance %p.\n", iface, instance );
 
@@ -556,18 +461,9 @@
     impl->IUISettings_iface.lpVtbl = &uisettings_vtbl;
     impl->IUISettings2_iface.lpVtbl = &uisettings2_vtbl;
     impl->IUISettings3_iface.lpVtbl = &uisettings3_vtbl;
-    impl->IUISettings4_iface.lpVtbl = &uisettings4_vtbl;
-    impl->IUISettings5_iface.lpVtbl = &uisettings5_vtbl;
-
-    if (FAILED(hr = weak_reference_source_init( &impl->weak_reference_source,
-                                                (IUnknown *)&impl->IUISettings_iface )))
-    {
-        *instance = NULL;
-        free( impl );
-        return hr;
-    }
-
-    *instance = (IInspectable *)&impl->IUISettings5_iface;
+    impl->ref = 1;
+
+    *instance = (IInspectable *)&impl->IUISettings3_iface;
     return S_OK;
 }
 
