diff --git a/dlls/qcap/v4l.c b/dlls/qcap/v4l.c
--- a/dlls/qcap/v4l.c
+++ b/dlls/qcap/v4l.c
@@ -27,9 +27,7 @@
 
 #include "config.h"
 
-#include <limits.h>
 #include <stdarg.h>
-#include <stdbool.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <fcntl.h>
@@ -143,7 +141,6 @@
         struct caps *caps = &device->caps[index];
 
         if (IsEqualGUID(&mt->formattype, &caps->media_type.formattype)
-                && IsEqualGUID(&mt->subtype, &caps->media_type.subtype)
                 && video_info->bmiHeader.biWidth == caps->video_info.bmiHeader.biWidth
                 && video_info->bmiHeader.biHeight == caps->video_info.bmiHeader.biHeight)
             return caps;
@@ -370,15 +367,13 @@
     return TRUE;
 }
 
-#define TICKSPERSEC ((LONGLONG)10000000)
-
 static void fill_caps(__u32 pixelformat, __u32 width, __u32 height,
-        LONGLONG max_frame_interval, LONGLONG min_frame_interval, struct caps *caps)
+        __u32 max_fps, __u32 min_fps, struct caps *caps)
 {
     LONG depth = 24;
 
     memset(caps, 0, sizeof(*caps));
-    caps->video_info.dwBitRate = width * height * depth * TICKSPERSEC / min_frame_interval;
+    caps->video_info.dwBitRate = width * height * depth * max_fps;
     caps->video_info.bmiHeader.biSize = sizeof(caps->video_info.bmiHeader);
     caps->video_info.bmiHeader.biWidth = width;
     caps->video_info.bmiHeader.biHeight = height;
@@ -396,16 +391,15 @@
     caps->media_type.cbFormat = sizeof(VIDEOINFOHEADER);
     /* We reallocate the caps array, so pbFormat has to be set after all caps
      * have been enumerated. */
-    caps->config.MaxFrameInterval = max_frame_interval;
-    caps->config.MinFrameInterval = min_frame_interval;
+    caps->config.MaxFrameInterval = 10000000 / max_fps;
+    caps->config.MinFrameInterval = 10000000 / min_fps;
     caps->config.MaxOutputSize.cx = width;
     caps->config.MaxOutputSize.cy = height;
     caps->config.MinOutputSize.cx = width;
     caps->config.MinOutputSize.cy = height;
     caps->config.guid = FORMAT_VideoInfo;
-    /* Dividing swaps max and min. */
-    caps->config.MinBitsPerSecond = width * height * depth * TICKSPERSEC / max_frame_interval;
-    caps->config.MaxBitsPerSecond = width * height * depth * TICKSPERSEC / min_frame_interval;
+    caps->config.MinBitsPerSecond = width * height * depth * min_fps;
+    caps->config.MaxBitsPerSecond = width * height * depth * max_fps;
     caps->pixelformat = pixelformat;
 }
 
@@ -426,37 +420,6 @@
     struct video_capture_device *device = get_device(params->device);
 
     *params->count = device->caps_count;
-    return S_OK;
-}
-
-static NTSTATUS v4l_device_get_frame_intervals( void *args )
-{
-    const struct get_frame_intervals_params *params = args;
-    struct video_capture_device *device = get_device(params->device);
-
-    struct v4l2_frmivalenum frmival =
-    {
-        .pixel_format = device->caps[params->caps_index].pixelformat,
-        .width = params->width,
-        .height = params->height,
-    };
-    __u32 count = 0;
-
-    for (frmival.index = 0; xioctl(device->fd, VIDIOC_ENUM_FRAMEINTERVALS, &frmival) != -1; ++frmival.index)
-    {
-        if (frmival.type == V4L2_FRMIVAL_TYPE_DISCRETE)
-        {
-            if (params->intervals)
-                params->intervals[count] = TICKSPERSEC * frmival.discrete.numerator / frmival.discrete.denominator;
-            ++count;
-        }
-        else if (frmival.type == V4L2_FRMIVAL_TYPE_STEPWISE || frmival.type == V4L2_FRMIVAL_TYPE_CONTINUOUS)
-        {
-            FIXME("Unhandled type %u.\n", frmival.type);
-        }
-    }
-
-    *params->count = count;
     return S_OK;
 }
 
@@ -538,8 +501,8 @@
     frmsize.pixel_format = V4L2_PIX_FMT_BGR24;
     while (xioctl(fd, VIDIOC_ENUM_FRAMESIZES, &frmsize) != -1)
     {
-        LONGLONG max_frame_interval = 0, min_frame_interval = LLONG_MAX;
         struct v4l2_frmivalenum frmival = {0};
+        __u32 max_fps = 30, min_fps = 30;
         struct caps *new_caps;
 
         frmival.pixel_format = format.fmt.pix.pixelformat;
@@ -559,27 +522,21 @@
             continue;
         }
 
-        for (frmival.index = 0; xioctl(fd, VIDIOC_ENUM_FRAMEINTERVALS, &frmival) != -1; ++frmival.index)
+        if (xioctl(fd, VIDIOC_ENUM_FRAMEINTERVALS, &frmival) != -1)
         {
             if (frmival.type == V4L2_FRMIVAL_TYPE_DISCRETE)
             {
-                LONGLONG interval = TICKSPERSEC * frmival.discrete.numerator / frmival.discrete.denominator;
-
-                max_frame_interval = max(max_frame_interval, interval);
-                min_frame_interval = min(min_frame_interval, interval);
+                max_fps = frmival.discrete.denominator / frmival.discrete.numerator;
+                min_fps = max_fps;
             }
             else if (frmival.type == V4L2_FRMIVAL_TYPE_STEPWISE
                     || frmival.type == V4L2_FRMIVAL_TYPE_CONTINUOUS)
             {
-                LONGLONG max_interval = TICKSPERSEC * frmival.stepwise.max.numerator / frmival.stepwise.max.denominator;
-                LONGLONG min_interval = TICKSPERSEC * frmival.stepwise.min.numerator / frmival.stepwise.min.denominator;
-
-                max_frame_interval = max(max_frame_interval, max_interval);
-                min_frame_interval = min(min_frame_interval, min_interval);
+                min_fps = frmival.stepwise.max.denominator / frmival.stepwise.max.numerator;
+                max_fps = frmival.stepwise.min.denominator / frmival.stepwise.min.numerator;
             }
         }
-
-        if (max_frame_interval < min_frame_interval)
+        else
             ERR("Failed to get fps: %s.\n", strerror(errno));
 
         new_caps = realloc(device->caps, (device->caps_count + 1) * sizeof(*device->caps));
@@ -587,7 +544,7 @@
             goto error;
         device->caps = new_caps;
         fill_caps(format.fmt.pix.pixelformat, frmsize.discrete.width, frmsize.discrete.height,
-                max_frame_interval, min_frame_interval, &device->caps[device->caps_count]);
+                max_fps, min_fps, &device->caps[device->caps_count]);
         device->caps_count++;
 
         frmsize.index++;
@@ -651,7 +608,6 @@
     v4l_device_get_media_type,
     v4l_device_get_caps,
     v4l_device_get_caps_count,
-    v4l_device_get_frame_intervals,
     v4l_device_get_prop_range,
     v4l_device_get_prop,
     v4l_device_set_prop,
@@ -843,31 +799,6 @@
     return v4l_device_get_caps_count( &params );
 }
 
-static NTSTATUS wow64_v4l_device_get_frame_intervals( void *args )
-{
-    struct
-    {
-        video_capture_device_t device;
-        unsigned int           caps_index;
-        unsigned int           width;
-        unsigned int           height;
-        PTR32                  count;
-        PTR32                  intervals;
-    } const *params32 = args;
-
-    struct get_frame_intervals_params params =
-    {
-        params32->device,
-        params32->caps_index,
-        params32->width,
-        params32->height,
-        ULongToPtr(params32->count),
-        ULongToPtr(params32->intervals),
-    };
-
-    return v4l_device_get_frame_intervals( &params );
-}
-
 static NTSTATUS wow64_v4l_device_get_prop_range( void *args )
 {
     struct
@@ -944,7 +875,6 @@
     wow64_v4l_device_get_media_type,
     wow64_v4l_device_get_caps,
     wow64_v4l_device_get_caps_count,
-    wow64_v4l_device_get_frame_intervals,
     wow64_v4l_device_get_prop_range,
     wow64_v4l_device_get_prop,
     v4l_device_set_prop,
