diff --git a/dlls/oledb32/tests/database.c b/dlls/oledb32/tests/database.c
--- a/dlls/oledb32/tests/database.c
+++ b/dlls/oledb32/tests/database.c
@@ -36,6 +36,8 @@
 
 #include "wine/test.h"
 
+DEFINE_GUID(CSLID_MSDAER, 0xc8b522cf,0x5cf3,0x11ce,0xad,0xe5,0x00,0xaa,0x00,0x44,0x77,0x3d);
+
 #define EXPECT_REF(obj,ref) _expect_ref((IUnknown*)obj, ref, __LINE__)
 static void _expect_ref(IUnknown* obj, ULONG ref, int line)
 {
@@ -161,6 +163,7 @@
     ok(IsEqualIID(&propsets->guidPropertySet, &DBPROPSET_DBINIT), "set guid %s\n", wine_dbgstr_guid(&propsets->guidPropertySet));
     ok(propsets->cProperties == 2, "got propcount %lu\n", propsets->cProperties);
 
+if (propsets->cProperties == 2) {
     ok(propsets->rgProperties[0].dwPropertyID == DBPROP_INIT_DATASOURCE, "got propid[0] %lu\n", propsets->rgProperties[0].dwPropertyID);
     ok(propsets->rgProperties[0].dwOptions == DBPROPOPTIONS_REQUIRED, "got options[0] %lu\n", propsets->rgProperties[0].dwOptions);
     ok(propsets->rgProperties[0].dwStatus == 0, "got status[0] %lu\n", propsets->rgProperties[0].dwStatus);
@@ -191,6 +194,7 @@
         ok(!wcscmp(V_BSTR(&propsets->rgProperties[1].vValue), L"a=1;b=2;c=3"), "got provider string %s\n",
            wine_dbgstr_variant(&propsets->rgProperties[1].vValue));
     }
+}
     return S_OK;
 }
 
@@ -320,7 +324,7 @@
     static const WCHAR *initial_catalog_prop = L"Data Source=initial_catalog_test;Initial Catalog=dummy_catalog";
     static const WCHAR *extended_prop = L"data source=dummy;Extended Properties=\"DRIVER=A Wine ODBC driver;UID=wine;\";";
     static const WCHAR *extended_prop2 = L"data source=\'dummy\';customprop=\'123.4\';";
-    static const WCHAR *multi_provider_prop_test = L"Data Source=provider_prop_test; a=1; b=2;c=3;";
+    static const WCHAR *multi_provider_prop_test = L"Data Source=provider_prop_test;a=1;b=2;c=3;";
     IDataInitialize *datainit = NULL;
     HRESULT hr;
 
@@ -361,15 +365,15 @@
     IErrorRecords *errrecs;
     IUnknown *unk = NULL, *unk2;
     DISPPARAMS dispparams;
+    DISPID dispid;
     DWORD context;
-    DISPID dispid;
     ULONG cnt = 0;
     VARIANT arg;
     HRESULT hr;
     GUID guid;
     BSTR str;
 
-    hr = CoCreateInstance(&CLSID_EXTENDEDERRORINFO, NULL, CLSCTX_INPROC_SERVER, &IID_IUnknown, (void**)&unk);
+    hr = CoCreateInstance(&CSLID_MSDAER, NULL, CLSCTX_INPROC_SERVER, &IID_IUnknown, (void**)&unk);
     ok(hr == S_OK, "got %08lx\n", hr);
 
     hr = IUnknown_QueryInterface(unk, &IID_IErrorInfo, (void**)&errorinfo);
@@ -415,6 +419,8 @@
     ok(hr == E_FAIL, "got %08lx\n", hr);
     ok(context == 0, "got %ld\n", context);
 
+    IErrorInfo_Release(errorinfo);
+
     hr = IErrorInfo_QueryInterface(errorinfo, &IID_ICreateErrorInfo, (void**)&createerror);
     ok(hr == E_NOINTERFACE, "got %08lx\n", hr);
 
@@ -439,49 +445,19 @@
 
     memset(&info, 0, sizeof(ERRORINFO));
     info.dwMinor = 1;
-    memcpy(&info.iid, &IID_IUnknown, sizeof(info.iid));
-    memcpy(&info.clsid, &IID_IDispatch, sizeof(info.clsid));
     memset(&info2, 0, sizeof(ERRORINFO));
     info2.dwMinor = 2;
-    memcpy(&info2.iid, &IID_IErrorInfo, sizeof(info2.iid));
     memset(&info3, 0, sizeof(ERRORINFO));
 
-    hr = IErrorRecords_AddErrorRecord(errrecs, NULL, IDENTIFIER_SDK_ERROR, NULL, NULL, 0);
+    hr = IErrorRecords_AddErrorRecord(errrecs, NULL, 268435456, NULL, NULL, 0);
     ok(hr == E_INVALIDARG, "got %08lx\n", hr);
 
-    /* Record that doesn't use lookup service. */
-    hr = IErrorRecords_AddErrorRecord(errrecs, &info, IDENTIFIER_SDK_ERROR, NULL, NULL, 0);
+    hr = IErrorRecords_AddErrorRecord(errrecs, &info, 1, NULL, NULL, 0);
     ok(hr == S_OK, "got %08lx\n", hr);
 
     hr = IErrorRecords_GetRecordCount(errrecs, &cnt);
     ok(hr == S_OK, "got %08lx\n", hr);
     ok(cnt == 1, "expected 1 got %ld\n", cnt);
-
-    memset(&guid, 0xac, sizeof(guid));
-    hr = IErrorInfo_GetGUID(errorinfo, &guid);
-    ok(hr == S_OK, "got %08lx\n", hr);
-    ok(IsEqualGUID(&guid, &info.iid), "got wrong guid\n");
-
-    str = (BSTR)0x1;
-    hr = IErrorInfo_GetSource(errorinfo, &str);
-    ok(hr == E_FAIL, "got %08lx\n", hr);
-    ok(str == NULL, "got %s\n", wine_dbgstr_w(str));
-
-    str = NULL;
-    hr = IErrorInfo_GetDescription(errorinfo, &str);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(!!str, "Expected error text.\n");
-    SysFreeString(str);
-
-    str = (BSTR)0x1;
-    hr = IErrorInfo_GetHelpFile(errorinfo, &str);
-    ok(hr == E_FAIL, "got %08lx\n", hr);
-    ok(str == NULL, "got %s\n", wine_dbgstr_w(str));
-
-    context = 1;
-    hr = IErrorInfo_GetHelpContext(errorinfo, &context);
-    ok(hr == E_FAIL, "got %08lx\n", hr);
-    ok(context == 0, "got %ld\n", context);
 
     /* Record does not contain custom error object. */
     unk2 = (void*)0xdeadbeef;
@@ -495,10 +471,6 @@
     hr = IErrorRecords_GetRecordCount(errrecs, &cnt);
     ok(hr == S_OK, "got %08lx\n", hr);
     ok(cnt == 2, "expected 2 got %ld\n", cnt);
-
-    hr = IErrorInfo_GetGUID(errorinfo, &guid);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(IsEqualGUID(&guid, &IID_IErrorInfo), "Unexpected guid.\n");
 
     hr = IErrorRecords_GetBasicErrorInfo(errrecs, 0, NULL);
     ok(hr == E_INVALIDARG, "got %08lx\n", hr);
@@ -543,68 +515,19 @@
     ok(dispparams.cArgs == 1, "Got argument count %u\n", dispparams.cArgs);
     ok(dispparams.cNamedArgs == 1, "Got named argument count %u\n", dispparams.cNamedArgs);
 
-    EXPECT_REF(errrecs, 3);
+    EXPECT_REF(errrecs, 2);
+    EXPECT_REF(errorinfo, 2);
+    hr = IErrorRecords_GetErrorInfo(errrecs, 0, 0, &errorinfo2);
+    ok(hr == S_OK, "got %08lx\n", hr);
+    ok(errorinfo == errorinfo2, "different object\n");
     EXPECT_REF(errorinfo, 3);
-    hr = IErrorRecords_GetErrorInfo(errrecs, 0, 0, &errorinfo2);
-    ok(hr == S_OK, "got %08lx\n", hr);
-    ok(errorinfo == errorinfo2, "different object\n");
-    EXPECT_REF(errorinfo, 4);
-    EXPECT_REF(errrecs, 4);
     IErrorInfo_Release(errorinfo2);
 
     free_dispparams(&dispparams);
     VariantClear(&arg);
 
-    IErrorInfo_Release(errorinfo);
     IErrorRecords_Release(errrecs);
     IUnknown_Release(unk);
-
-    /* Shift in record indexing. Same returned IErrorInfo instance refers to a live record collection,
-       not a copy. */
-    hr = CoCreateInstance(&CLSID_EXTENDEDERRORINFO, NULL, CLSCTX_INPROC_SERVER, &IID_IErrorRecords, (void **)&errrecs);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    memset(&info, 0, sizeof(info));
-    memcpy(&info.iid, &IID_IUnknown, sizeof(info.iid));
-    hr = IErrorRecords_AddErrorRecord(errrecs, &info, IDENTIFIER_SDK_ERROR, NULL, NULL, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IErrorRecords_GetErrorInfo(errrecs, 0, 0, &errorinfo);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IErrorInfo_GetGUID(errorinfo, &guid);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(IsEqualGUID(&guid, &IID_IUnknown), "Unexpected guid %s.\n", wine_dbgstr_guid(&guid));
-
-    memset(&info, 0, sizeof(info));
-    memcpy(&info.iid, &IID_IDispatch, sizeof(info.iid));
-    hr = IErrorRecords_AddErrorRecord(errrecs, &info, IDENTIFIER_SDK_ERROR, NULL, NULL, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IErrorRecords_GetErrorInfo(errrecs, 1, 0, &errorinfo2);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IErrorInfo_GetGUID(errorinfo2, &guid);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(IsEqualGUID(&guid, &IID_IUnknown), "Unexpected guid %s.\n", wine_dbgstr_guid(&guid));
-
-    hr = IErrorInfo_GetGUID(errorinfo, &guid);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(IsEqualGUID(&guid, &IID_IDispatch), "Unexpected guid %s.\n", wine_dbgstr_guid(&guid));
-
-    memset(&info, 0, sizeof(info));
-    memcpy(&info.iid, &IID_IErrorInfo, sizeof(info.iid));
-    hr = IErrorRecords_AddErrorRecord(errrecs, &info, IDENTIFIER_SDK_ERROR, NULL, NULL, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IErrorInfo_GetGUID(errorinfo2, &guid);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(IsEqualGUID(&guid, &IID_IDispatch), "Unexpected guid %s.\n", wine_dbgstr_guid(&guid));
-
-    hr = IErrorInfo_GetGUID(errorinfo, &guid);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(IsEqualGUID(&guid, &IID_IErrorInfo), "Unexpected guid %s.\n", wine_dbgstr_guid(&guid));
-
-    IErrorInfo_Release(errorinfo2);
-    IErrorInfo_Release(errorinfo);
-    IErrorRecords_Release(errrecs);
 }
 
 #define expect_initstring(a, b, c) _expect_initstring(__LINE__, a, b, c)
