diff --git a/dlls/msv1_0/main.c b/dlls/msv1_0/main.c
--- a/dlls/msv1_0/main.c
+++ b/dlls/msv1_0/main.c
@@ -43,6 +43,11 @@
 static ULONG ntlm_package_id;
 static LSA_DISPATCH_TABLE lsa_dispatch;
 
+static NTSTATUS ntlm_check_version(void)
+{
+    return WINE_UNIX_CALL( unix_check_version, NULL );
+}
+
 static void ntlm_cleanup( struct ntlm_ctx *ctx )
 {
     WINE_UNIX_CALL( unix_cleanup, ctx );
@@ -108,6 +113,12 @@
     TRACE( "%#lx, %p, %s, %s, %p\n", package_id, dispatch, debugstr_as(database), debugstr_as(confidentiality),
            package_name );
 
+    if (ntlm_check_version())
+    {
+        ERR( "no NTLM support, expect problems\n" );
+        return STATUS_UNSUCCESSFUL;
+    }
+
     if (!(str = dispatch->AllocateLsaHeap( sizeof(*str) + sizeof("NTLM" )))) return STATUS_NO_MEMORY;
     ptr = (char *)(str + 1);
     memcpy( ptr, "NTLM", sizeof("NTLM") );
@@ -124,6 +135,12 @@
                                          LSA_SECPKG_FUNCTION_TABLE *lsa_function_table )
 {
     TRACE( "%#Ix, %p, %p\n", package_id, params, lsa_function_table );
+
+    if (ntlm_check_version())
+    {
+        ERR( "no NTLM support, expect problems\n" );
+        return STATUS_UNSUCCESSFUL;
+    }
     return STATUS_SUCCESS;
 }
 
@@ -1322,11 +1339,11 @@
     memcpy( digest, outer_ctx.digest, 16 );
 }
 
-static SECURITY_STATUS create_signature( struct ntlm_ctx *ctx, unsigned int flags, SecBufferDesc *msg,
-                                         SecBuffer *sig_buf, enum sign_direction dir, BOOL encrypt )
+static SECURITY_STATUS create_signature( struct ntlm_ctx *ctx, unsigned int flags, SecBufferDesc *msg, int idx,
+                                         enum sign_direction dir, BOOL encrypt )
 {
     unsigned int i, sign_version = 1;
-    char *sig = sig_buf->pvBuffer;
+    char *sig = msg->pBuffers[idx].pvBuffer;
 
     if (flags & FLAG_NEGOTIATE_NTLM2 && flags & FLAG_NEGOTIATE_SIGN)
     {
@@ -1377,7 +1394,7 @@
         memcpy( sig + 4, digest, 8 );
         memcpy( sig + 12, seq_no, 4 );
 
-        sig_buf->cbBuffer = 16;
+        msg->pBuffers[idx].cbBuffer = 16;
         return SEC_E_OK;
     }
 
@@ -1413,9 +1430,9 @@
     if (flags & FLAG_NEGOTIATE_ALWAYS_SIGN || !flags)
     {
         /* create dummy signature */
-        memset( sig_buf->pvBuffer, 0, 16 );
-        memset( sig_buf->pvBuffer, 1, 1 );
-        sig_buf->cbBuffer = 16;
+        memset( msg->pBuffers[idx].pvBuffer, 0, 16 );
+        memset( msg->pBuffers[idx].pvBuffer, 1, 1 );
+        msg->pBuffers[idx].cbBuffer = 16;
         return SEC_E_OK;
     }
 
@@ -1436,13 +1453,13 @@
         return SEC_E_INVALID_TOKEN;
     if (msg->pBuffers[idx].cbBuffer < 16) return SEC_E_BUFFER_TOO_SMALL;
 
-    return create_signature( ctx, ctx->flags, msg, &msg->pBuffers[idx], SIGN_SEND, TRUE );
-}
-
-static NTSTATUS verify_signature( struct ntlm_ctx *ctx, unsigned int flags, SecBufferDesc *msg, SecBuffer *sig_buf )
+    return create_signature( ctx, ctx->flags, msg, idx, SIGN_SEND, TRUE );
+}
+
+static NTSTATUS verify_signature( struct ntlm_ctx *ctx, unsigned int flags, SecBufferDesc *msg, int idx )
 {
     NTSTATUS status;
-    unsigned int i, sig_idx = 0;
+    unsigned int i;
     SecBufferDesc desc;
     SecBuffer *buf;
     char sig[16];
@@ -1454,12 +1471,11 @@
 
     for (i = 0; i < msg->cBuffers; i++)
     {
-        if (msg->pBuffers[i].BufferType == SECBUFFER_TOKEN || msg->pBuffers[i].BufferType == SECBUFFER_STREAM)
+        if (msg->pBuffers[i].BufferType == SECBUFFER_TOKEN)
         {
             buf[i].BufferType = SECBUFFER_TOKEN;
             buf[i].cbBuffer   = 16;
             buf[i].pvBuffer   = sig;
-            sig_idx = i;
         }
         else
         {
@@ -1469,9 +1485,9 @@
         }
     }
 
-    if ((status = create_signature( ctx, flags, &desc, &buf[sig_idx], SIGN_RECV, TRUE )) == SEC_E_OK)
-    {
-        if (memcmp( (char *)buf[sig_idx].pvBuffer + 8, (char *)sig_buf->pvBuffer + 8, 8 ))
+    if ((status = create_signature( ctx, flags, &desc, idx, SIGN_RECV, TRUE )) == SEC_E_OK)
+    {
+        if (memcmp( (char *)buf[idx].pvBuffer + 8, (char *)msg->pBuffers[idx].pvBuffer + 8, 8 ))
             status = SEC_E_MESSAGE_ALTERED;
     }
 
@@ -1492,7 +1508,7 @@
         return SEC_E_INVALID_TOKEN;
     if (msg->pBuffers[idx].cbBuffer < 16) return SEC_E_BUFFER_TOO_SMALL;
 
-    return verify_signature( ctx, ctx->flags, msg, &msg->pBuffers[idx] );
+    return verify_signature( ctx, ctx->flags, msg, idx );
 }
 
 static NTSTATUS NTAPI ntlm_SpSealMessage( LSA_SEC_HANDLE handle, ULONG qop, SecBufferDesc *msg, ULONG msg_seq_no )
@@ -1515,7 +1531,7 @@
     ctx = (struct ntlm_ctx *)handle;
     if (ctx->flags & FLAG_NEGOTIATE_NTLM2 && ctx->flags & FLAG_NEGOTIATE_SEAL)
     {
-        create_signature( ctx, ctx->flags, msg, &msg->pBuffers[token_idx], SIGN_SEND, FALSE );
+        create_signature( ctx, ctx->flags, msg, token_idx, SIGN_SEND, FALSE );
 
         arc4_process( &ctx->crypt.ntlm2.send_arc4info, msg->pBuffers[data_idx].pvBuffer,
                       msg->pBuffers[data_idx].cbBuffer );
@@ -1526,7 +1542,7 @@
     {
         char *sig = msg->pBuffers[token_idx].pvBuffer;
 
-        create_signature( ctx, ctx->flags | FLAG_NEGOTIATE_SIGN, msg, &msg->pBuffers[token_idx], SIGN_SEND, FALSE );
+        create_signature( ctx, ctx->flags | FLAG_NEGOTIATE_SIGN, msg, token_idx, SIGN_SEND, FALSE );
 
         arc4_process( &ctx->crypt.ntlm.arc4info, msg->pBuffers[data_idx].pvBuffer, msg->pBuffers[data_idx].cbBuffer );
         arc4_process( &ctx->crypt.ntlm.arc4info, sig + 4, 12 );
@@ -1539,8 +1555,7 @@
 
 static NTSTATUS NTAPI ntlm_SpUnsealMessage( LSA_SEC_HANDLE handle, SecBufferDesc *msg, ULONG msg_seq_no, ULONG *qop )
 {
-    int data_idx, stream_idx, token_idx;
-    SecBuffer token_buf;
+    int token_idx, data_idx;
     struct ntlm_ctx *ctx;
 
     TRACE( "%#Ix, %p, %lu, %p\n", handle, msg, msg_seq_no, qop );
@@ -1548,31 +1563,11 @@
 
     if (!handle) return SEC_E_INVALID_HANDLE;
 
-    if (!msg || !msg->pBuffers || msg->cBuffers < 2) return SEC_E_INVALID_TOKEN;
-
-    data_idx = get_buffer_index( msg, SECBUFFER_DATA );
-    stream_idx = get_buffer_index( msg, SECBUFFER_STREAM );
-    token_idx = get_buffer_index( msg, SECBUFFER_TOKEN );
-
-    if ((token_idx == -1 && stream_idx == -1) || (stream_idx != -1 && token_idx != -1) || data_idx == -1)
-        return SEC_E_INVALID_TOKEN;
-
-    if (stream_idx != -1)
-    {
-        if (msg->pBuffers[stream_idx].cbBuffer < 16) return SEC_E_BUFFER_TOO_SMALL;
-        token_buf.BufferType = SECBUFFER_TOKEN;
-        token_buf.cbBuffer = 16;
-        token_buf.pvBuffer = msg->pBuffers[stream_idx].pvBuffer;
-
-        /* native decrypts in-place even when an appropriately sized data buffer is supplied */
-        msg->pBuffers[data_idx].pvBuffer = (char *)msg->pBuffers[stream_idx].pvBuffer + 16;
-        msg->pBuffers[data_idx].cbBuffer = msg->pBuffers[stream_idx].cbBuffer - 16;
-    }
-    else
-    {
-        if (msg->pBuffers[token_idx].cbBuffer < 16) return SEC_E_BUFFER_TOO_SMALL;
-        token_buf = msg->pBuffers[token_idx];
-    }
+    if (!msg || !msg->pBuffers || msg->cBuffers < 2 ||
+        (token_idx = get_buffer_index( msg, SECBUFFER_TOKEN )) == -1 ||
+        (data_idx = get_buffer_index( msg, SECBUFFER_DATA )) == -1) return SEC_E_INVALID_TOKEN;
+
+    if (msg->pBuffers[token_idx].cbBuffer < 16) return SEC_E_BUFFER_TOO_SMALL;
 
     ctx = (struct ntlm_ctx *)handle;
     if (ctx->flags & FLAG_NEGOTIATE_NTLM2 && ctx->flags & FLAG_NEGOTIATE_SEAL)
@@ -1584,7 +1579,7 @@
 
     /* make sure we use a session key for the signature check, SealMessage always does that,
        even in the dummy case */
-    return verify_signature( ctx, ctx->flags | FLAG_NEGOTIATE_SIGN, msg, &token_buf );
+    return verify_signature( ctx, ctx->flags | FLAG_NEGOTIATE_SIGN, msg, token_idx );
 }
 
 static SECPKG_USER_FUNCTION_TABLE ntlm_user_table =
