diff --git a/dlls/windowscodecs/wincodecs_private.h b/dlls/windowscodecs/wincodecs_private.h
--- a/dlls/windowscodecs/wincodecs_private.h
+++ b/dlls/windowscodecs/wincodecs_private.h
@@ -19,12 +19,16 @@
 #ifndef WINCODECS_PRIVATE_H
 #define WINCODECS_PRIVATE_H
 
-#include <stdbool.h>
-
 #include "wincodec.h"
 #include "wincodecsdk.h"
 
 #include "wine/debug.h"
+
+DEFINE_GUID(CLSID_WineTgaDecoder, 0xb11fc79a,0x67cc,0x43e6,0xa9,0xce,0xe3,0xd5,0x49,0x45,0xd3,0x04);
+
+DEFINE_GUID(GUID_WineContainerFormatTga, 0x0c44fda1,0xa5c5,0x4298,0x96,0x85,0x47,0x3f,0xc1,0x7c,0xd3,0x22);
+
+DEFINE_GUID(GUID_VendorWine, 0xddf46da1,0x7dc1,0x404e,0x98,0xf2,0xef,0xa4,0x8d,0xfc,0x95,0x0a);
 
 extern IID IID_IMILBitmap;
 extern IID IID_IMILBitmapSource;
@@ -158,8 +162,6 @@
 extern HRESULT ColorTransform_Create(IWICColorTransform **transform);
 extern HRESULT BitmapClipper_Create(IWICBitmapClipper **clipper);
 
-extern HRESULT create_stream_wrapper(IStream *input, ULONG offset, IStream **wrapper);
-
 extern HRESULT copy_pixels(UINT bpp, const BYTE *srcbuffer,
     UINT srcwidth, UINT srcheight, INT srcstride,
     const WICRect *rc, UINT dststride, UINT dstbuffersize, BYTE *dstbuffer);
@@ -192,20 +194,6 @@
 extern void BmpDecoder_GetWICDecoder(BmpDecoder *This, IWICBitmapDecoder **ppDecoder);
 extern void BmpDecoder_FindIconMask(BmpDecoder *This, ULONG *mask_offset, int *topdown);
 
-static inline HRESULT init_propvar_from_string(const WCHAR *str, PROPVARIANT *var)
-{
-    size_t size = (wcslen(str) + 1) * sizeof(*str);
-    WCHAR *s;
-
-    if (!(s = CoTaskMemAlloc(size)))
-        return E_OUTOFMEMORY;
-    memcpy(s, str, size);
-
-    var->pwszVal = s;
-    var->vt = VT_LPWSTR;
-    return S_OK;
-}
-
 typedef struct _MetadataItem
 {
     PROPVARIANT schema;
@@ -213,72 +201,35 @@
     PROPVARIANT value;
 } MetadataItem;
 
-enum metadatahandler_flags
-{
-    METADATAHANDLER_IS_WRITER = 0x1,
-    METADATAHANDLER_FIXED_ITEMS = 0x2, /* Items cannot be added or removed. */
-};
-
-typedef struct MetadataHandler MetadataHandler;
-
 typedef struct _MetadataHandlerVtbl
 {
-    DWORD flags;
+    int is_writer;
     const CLSID *clsid;
-    HRESULT (*fnLoad)(MetadataHandler *handler, IStream *stream, const GUID *preferred_vendor,
-        DWORD persist_options);
-    HRESULT (*fnCreate)(MetadataHandler *handler);
+    HRESULT (*fnLoad)(IStream *stream, const GUID *preferred_vendor,
+        DWORD persist_options, MetadataItem **items, DWORD *item_count);
+    HRESULT (*fnSave)(IStream *stream, DWORD persist_options,
+        const MetadataItem *items, DWORD item_count);
+    HRESULT (*fnGetSizeMax)(const MetadataItem *items, DWORD item_count,
+        ULARGE_INTEGER *size);
 } MetadataHandlerVtbl;
 
-typedef struct MetadataHandler
-{
-    IWICMetadataWriter IWICMetadataWriter_iface;
-    LONG ref;
-    IWICPersistStream IWICPersistStream_iface;
-    IWICStreamProvider IWICStreamProvider_iface;
-    const MetadataHandlerVtbl *vtable;
-    MetadataItem *items;
-    DWORD item_count;
-    DWORD persist_options;
-    IStream *stream;
-    ULARGE_INTEGER origin;
-    CRITICAL_SECTION lock;
-} MetadataHandler;
-
 extern HRESULT MetadataReader_Create(const MetadataHandlerVtbl *vtable, REFIID iid, void** ppv);
-extern void MetadataHandler_FreeItems(MetadataHandler *handler);
-extern void clear_metadata_item(MetadataItem *item);
 
 extern HRESULT UnknownMetadataReader_CreateInstance(REFIID iid, void** ppv);
-extern HRESULT UnknownMetadataWriter_CreateInstance(REFIID iid, void** ppv);
 extern HRESULT IfdMetadataReader_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT IfdMetadataWriter_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT GpsMetadataReader_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT GpsMetadataWriter_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT ExifMetadataReader_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT ExifMetadataWriter_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT App1MetadataReader_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT App1MetadataWriter_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT PngBkgdReader_CreateInstance(REFIID iid, void** ppv);
-extern HRESULT PngBkgdWriter_CreateInstance(REFIID iid, void** ppv);
 extern HRESULT PngChrmReader_CreateInstance(REFIID iid, void** ppv);
 extern HRESULT PngGamaReader_CreateInstance(REFIID iid, void** ppv);
 extern HRESULT PngHistReader_CreateInstance(REFIID iid, void** ppv);
 extern HRESULT PngTextReader_CreateInstance(REFIID iid, void** ppv);
 extern HRESULT PngTimeReader_CreateInstance(REFIID iid, void** ppv);
-extern HRESULT PngTimeWriter_CreateInstance(REFIID iid, void** ppv);
 extern HRESULT LSDReader_CreateInstance(REFIID iid, void **ppv);
 extern HRESULT IMDReader_CreateInstance(REFIID iid, void **ppv);
 extern HRESULT GCEReader_CreateInstance(REFIID iid, void **ppv);
 extern HRESULT APEReader_CreateInstance(REFIID iid, void **ppv);
 extern HRESULT GifCommentReader_CreateInstance(REFIID iid, void **ppv);
-extern HRESULT MetadataQueryReader_CreateInstanceFromBlockReader(IWICMetadataBlockReader *, IWICMetadataQueryReader **);
-extern HRESULT MetadataQueryWriter_CreateInstanceFromBlockWriter(IWICMetadataBlockWriter *, IWICMetadataQueryWriter **);
-extern HRESULT MetadataQueryReader_CreateInstance(IWICMetadataReader *, IWICMetadataQueryReader **);
-extern HRESULT MetadataQueryWriter_CreateInstance(IWICMetadataWriter *, IWICMetadataQueryWriter **);
+extern HRESULT MetadataQueryReader_CreateInstance(IWICMetadataBlockReader *, const WCHAR *, IWICMetadataQueryReader **);
+extern HRESULT MetadataQueryWriter_CreateInstance(IWICMetadataBlockWriter *, const WCHAR *, IWICMetadataQueryWriter **);
 extern HRESULT stream_initialize_from_filehandle(IWICStream *iface, HANDLE hfile);
-
-extern bool wincodecs_array_reserve(void **elements, size_t *capacity, size_t count, size_t size);
 
 static inline const char *debug_wic_rect(const WICRect *rect)
 {
@@ -299,12 +250,8 @@
     CLSID clsid;
 };
 
-enum decoder_flags
-{
-    DECODER_FLAGS_CAPABILITY_MASK = 0x1f,
-    DECODER_FLAGS_UNSUPPORTED_COLOR_CONTEXT = 0x80000000,
-    DECODER_FLAGS_METADATA_AT_DECODER = 0x40000000,
-};
+#define DECODER_FLAGS_CAPABILITY_MASK 0x1f
+#define DECODER_FLAGS_UNSUPPORTED_COLOR_CONTEXT 0x80000000
 
 struct decoder_stat
 {
@@ -323,14 +270,9 @@
     WICColor palette[256];
 };
 
-enum decoder_block_options
-{
-    DECODER_BLOCK_OPTION_MASK = 0x0001000F,
-    DECODER_BLOCK_FULL_STREAM = 0x80000000,
-    DECODER_BLOCK_READER_CLSID = 0x40000000,
-    DECODER_BLOCK_OFFSET_IS_PTR = 0x20000000,
-};
-
+#define DECODER_BLOCK_OPTION_MASK 0x0001000F
+#define DECODER_BLOCK_FULL_STREAM 0x80000000
+#define DECODER_BLOCK_READER_CLSID 0x40000000
 struct decoder_block
 {
     ULONGLONG offset;
@@ -348,7 +290,6 @@
 {
     HRESULT (CDECL *initialize)(struct decoder* This, IStream *stream, struct decoder_stat *st);
     HRESULT (CDECL *get_frame_info)(struct decoder* This, UINT frame, struct decoder_frame *info);
-    HRESULT (CDECL *get_decoder_palette)(struct decoder* This, UINT frame, WICColor *colors, UINT *num_colors);
     HRESULT (CDECL *copy_pixels)(struct decoder* This, UINT frame, const WICRect *prc,
         UINT stride, UINT buffersize, BYTE *buffer);
     HRESULT (CDECL *get_metadata_blocks)(struct decoder* This, UINT frame, UINT *count,
@@ -366,7 +307,6 @@
 HRESULT CDECL decoder_create(const CLSID *decoder_clsid, struct decoder_info *info, struct decoder **result);
 HRESULT CDECL decoder_initialize(struct decoder *This, IStream *stream, struct decoder_stat *st);
 HRESULT CDECL decoder_get_frame_info(struct decoder* This, UINT frame, struct decoder_frame *info);
-HRESULT CDECL decoder_get_decoder_palette(struct decoder* This, UINT frame, WICColor *colors, UINT *num_colors);
 HRESULT CDECL decoder_copy_pixels(struct decoder* This, UINT frame, const WICRect *prc,
     UINT stride, UINT buffersize, BYTE *buffer);
 HRESULT CDECL decoder_get_metadata_blocks(struct decoder* This, UINT frame, UINT *count,
@@ -374,15 +314,6 @@
 HRESULT CDECL decoder_get_color_context(struct decoder* This, UINT frame, UINT num,
     BYTE **data, DWORD *datasize);
 void CDECL decoder_destroy(struct decoder *This);
-
-HRESULT create_metadata_reader(REFGUID format, const GUID *vendor, DWORD options, IStream *stream,
-        IWICMetadataReader **reader);
-HRESULT create_metadata_writer(REFGUID format, const GUID *vendor, DWORD options,
-        IWICMetadataWriter **writer);
-HRESULT create_metadata_query_writer_from_reader(IWICMetadataQueryReader *query_reader, const GUID *vendor,
-        IWICMetadataQueryWriter **query_writer);
-HRESULT create_metadata_writer_from_reader(IWICMetadataReader *reader, const GUID *vendor,
-        IWICMetadataWriter **writer);
 
 struct encoder_funcs;
 
@@ -467,4 +398,9 @@
 extern HRESULT CommonEncoder_CreateInstance(struct encoder *encoder,
     const struct encoder_info *encoder_info, REFIID iid, void** ppv);
 
+#ifdef _WIN64
+#undef setjmp
+#define setjmp(buf) _setjmpex(buf, NULL)
+#endif
+
 #endif /* WINCODECS_PRIVATE_H */
