diff --git a/dlls/ucrtbase/tests/cpp.c b/dlls/ucrtbase/tests/cpp.c
--- a/dlls/ucrtbase/tests/cpp.c
+++ b/dlls/ucrtbase/tests/cpp.c
@@ -24,38 +24,9 @@
 
 #include <windef.h>
 #include <winbase.h>
-#include <winternl.h>
-#include <rtlsupportapi.h>
 #include <verrsrc.h>
 #include <dbghelp.h>
-#include <unknwn.h>
 #include "wine/test.h"
-
-#define DEFINE_EXPECT(func) \
-    static BOOL expect_ ## func = FALSE, called_ ## func = FALSE
-
-#define SET_EXPECT(func) \
-    expect_ ## func = TRUE
-
-#define CHECK_EXPECT2(func) \
-    do { \
-        ok(expect_ ##func, "unexpected call " #func "\n"); \
-        called_ ## func = TRUE; \
-    }while(0)
-
-#define CHECK_EXPECT(func) \
-    do { \
-        CHECK_EXPECT2(func); \
-        expect_ ## func = FALSE; \
-    }while(0)
-
-#define CHECK_CALLED(func) \
-    do { \
-        ok(called_ ## func, "expected " #func "\n"); \
-        expect_ ## func = called_ ## func = FALSE; \
-    }while(0)
-
-DEFINE_EXPECT(iunknown_except_Release);
 
 typedef unsigned char MSVCRT_bool;
 
@@ -75,8 +46,6 @@
     SLIST_ENTRY entry;
     char name[1];
 } type_info_list;
-
-void CDECL __DestructExceptionObject(EXCEPTION_RECORD*);
 
 static void* (CDECL *p_malloc)(size_t);
 static void (CDECL *p___std_exception_copy)(const __std_exception_data*, __std_exception_data*);
@@ -253,10 +222,6 @@
            UNDNAME_NO_ARGUMENTS},
 /*   8 */ {"??$foo@J_W$$T@bar@@YAJQB_W$$THQAUgod@@@Z",
            "long __cdecl bar::foo<long,wchar_t,std::nullptr_t>(wchar_t const * const,std::nullptr_t,int,struct god * const)"},
-/*   9 */ {"?AU?$DeprecatedRAIIContextManager@UEnablePythonDispatcher@?A0x1272d024@@$$V@impl@torch@@",
-           "struct torch::impl::DeprecatedRAIIContextManager<struct `anonymous namespace'::EnablePythonDispatcher>",
-           NULL,
-           UNDNAME_NO_ARGUMENTS},
 
     };
     unsigned i;
@@ -270,147 +235,10 @@
     }
 }
 
-static HRESULT WINAPI iunknown_except_QueryInterface(IUnknown *iface,
-        REFIID riid, void**ppv)
-{
-    ok(0, "unexpected call\n");
-    return E_NOTIMPL;
-}
-
-static ULONG WINAPI iunknown_except_AddRef(IUnknown *iface)
-{
-    ok(0, "unexpected call\n");
-    return 2;
-}
-
-static ULONG WINAPI iunknown_except_Release(IUnknown *iface)
-{
-    CHECK_EXPECT(iunknown_except_Release);
-    return 1;
-}
-
-static IUnknownVtbl iunknown_except_vtbl = {
-    iunknown_except_QueryInterface,
-    iunknown_except_AddRef,
-    iunknown_except_Release
-};
-
-static IUnknown iunknown_except = { &iunknown_except_vtbl };
-
-static void test___DestructExceptionObject(void)
-{
-    IUnknown *piunk = &iunknown_except;
-    EXCEPTION_RECORD rec;
-    struct
-    {
-        UINT flags;
-        UINT arch_specific_data[8];
-    } info = { 0x10 };
-
-    __DestructExceptionObject(NULL);
-
-    memset(&rec, 0, sizeof(rec));
-    rec.ExceptionCode = 0xe06d7363;
-#ifdef __i386__
-    rec.NumberParameters = 3;
-#else
-    rec.NumberParameters = 4;
-#endif
-    rec.ExceptionInformation[0] = 0x19930520;
-    rec.ExceptionInformation[1] = (ULONG_PTR)&piunk;
-    rec.ExceptionInformation[2] = (ULONG_PTR)&info;
-    rec.ExceptionInformation[3] = (ULONG_PTR)GetModuleHandleA(NULL);
-    SET_EXPECT(iunknown_except_Release);
-    __DestructExceptionObject(&rec);
-    CHECK_CALLED(iunknown_except_Release);
-}
-
-#ifdef __x86_64__
-EXCEPTION_DISPOSITION WINAPI __C_specific_handler(EXCEPTION_RECORD*,void*,CONTEXT*,DISPATCHER_CONTEXT*);
-DEFINE_EXPECT(filter);
-static CONTEXT ctx;
-
-static ULONG WINAPI iunknown_except2_Release(IUnknown *iface)
-{
-    CHECK_EXPECT(iunknown_except_Release);
-    NtContinue(&ctx, FALSE);
-    return 1;
-}
-
-static IUnknownVtbl iunknown_except2_vtbl = {
-    iunknown_except_QueryInterface,
-    iunknown_except_AddRef,
-    iunknown_except2_Release
-};
-
-static IUnknown iunknown_except2 = { &iunknown_except2_vtbl };
-
-static LONG WINAPI filter(struct _EXCEPTION_POINTERS *ep, void *frame)
-{
-    CHECK_EXPECT(filter);
-
-    ok(frame == (void*)0x1234, "frame = %p\n", (void*)frame);
-    SET_EXPECT(iunknown_except_Release);
-    return EXCEPTION_EXECUTE_HANDLER;
-}
-
-static void test___C_specific_handler(void)
-{
-    struct
-    {
-        UINT flags;
-        UINT arch_specific_data[8];
-    } info = { 0x10 };
-    IUnknown *piunk = &iunknown_except2;
-    DISPATCHER_CONTEXT dispatch;
-    SCOPE_TABLE scope_table;
-    EXCEPTION_RECORD rec;
-    CONTEXT context;
-    LONG pass = 0;
-
-    InterlockedIncrement(&pass);
-    RtlCaptureContext(&ctx);
-    if (InterlockedIncrement(&pass) == 3)
-    {
-        CHECK_CALLED(filter);
-        CHECK_CALLED(iunknown_except_Release);
-        return;
-    }
-
-    memset(&rec, 0, sizeof(rec));
-    rec.ExceptionCode = 0xe06d7363;
-    rec.NumberParameters = 4;
-    rec.ExceptionInformation[0] = 0x19930520;
-    rec.ExceptionInformation[1] = (ULONG_PTR)&piunk;
-    rec.ExceptionInformation[2] = (ULONG_PTR)&info;
-    rec.ExceptionInformation[3] = (ULONG_PTR)GetModuleHandleA(NULL);
-
-    memset(&dispatch, 0, sizeof(dispatch));
-    dispatch.ImageBase = (ULONG_PTR)GetModuleHandleA(NULL);
-    dispatch.ControlPc = dispatch.ImageBase + 0x200;
-    dispatch.HandlerData = &scope_table;
-    dispatch.ContextRecord = &context;
-    scope_table.Count = 1;
-    scope_table.ScopeRecord[0].BeginAddress = 0x200;
-    scope_table.ScopeRecord[0].EndAddress = 0x400;
-    scope_table.ScopeRecord[0].HandlerAddress = (ULONG_PTR)filter - dispatch.ImageBase;
-    scope_table.ScopeRecord[0].JumpTarget = 1;
-    memset(&context, 0, sizeof(context));
-
-    SET_EXPECT(filter);
-    __C_specific_handler(&rec, (void*)0x1234, &context, &dispatch);
-    ok(0, "should not be reached\n");
-}
-#endif
-
 START_TEST(cpp)
 {
     if (!init()) return;
     test___std_exception();
     test___std_type_info();
     test___unDName();
-    test___DestructExceptionObject();
-#ifdef __x86_64__
-    test___C_specific_handler();
-#endif
-}
+}
