diff --git a/dlls/comctl32/trackbar.c b/dlls/comctl32/trackbar.c
--- a/dlls/comctl32/trackbar.c
+++ b/dlls/comctl32/trackbar.c
@@ -32,6 +32,8 @@
 #include "winuser.h"
 #include "winnls.h"
 #include "commctrl.h"
+#include "uxtheme.h"
+#include "vssym32.h"
 #include "wine/debug.h"
 
 #include "comctl32.h"
@@ -94,6 +96,8 @@
 #define TIC_SELECTIONMARKMAX    0x80
 #define TIC_SELECTIONMARKMIN    0x100
 #define TIC_SELECTIONMARK       (TIC_SELECTIONMARKMAX | TIC_SELECTIONMARKMIN)
+
+static const WCHAR themeClass[] = L"Trackbar";
 
 static inline int 
 notify_customdraw (const TRACKBAR_INFO *infoPtr, NMCUSTOMDRAW *pnmcd, int stage)
@@ -459,25 +463,22 @@
 TRACKBAR_DrawChannel (const TRACKBAR_INFO *infoPtr, HDC hdc)
 {
     RECT rcChannel = infoPtr->rcChannel;
-
-#if __WINE_COMCTL32_VERSION == 6
     HTHEME theme = GetWindowTheme (infoPtr->hwndSelf);
 
     if (theme)
     {
-        int part = (infoPtr->dwStyle & TBS_VERT) ? TKP_TRACKVERT : TKP_TRACK;
-        DrawThemeBackground (theme, hdc, part, TKS_NORMAL, &rcChannel, 0);
-        return;
-    }
-#endif
-
-    DrawEdge (hdc, &rcChannel, EDGE_SUNKEN, BF_RECT | BF_ADJUST);
-    if (infoPtr->dwStyle & TBS_ENABLESELRANGE)
+        DrawThemeBackground (theme, hdc, 
+            (infoPtr->dwStyle & TBS_VERT) ?
+                TKP_TRACKVERT : TKP_TRACK, TKS_NORMAL, &rcChannel, 0);
+    }
+    else
     {
-        /* fill the channel */
-        FillRect (hdc, &rcChannel, GetStockObject (WHITE_BRUSH));
-        if (TRACKBAR_HasSelection (infoPtr))
-            FillRect (hdc, &infoPtr->rcSelection, GetSysColorBrush (COLOR_HIGHLIGHT));
+        DrawEdge (hdc, &rcChannel, EDGE_SUNKEN, BF_RECT | BF_ADJUST);
+        if (infoPtr->dwStyle & TBS_ENABLESELRANGE) {		 /* fill the channel */
+            FillRect (hdc, &rcChannel, GetStockObject(WHITE_BRUSH));
+            if (TRACKBAR_HasSelection(infoPtr))
+                FillRect (hdc, &infoPtr->rcSelection, GetSysColorBrush(COLOR_HIGHLIGHT));
+        }
     }
 }
 
@@ -562,37 +563,21 @@
         TRACKBAR_DrawOneTic (infoPtr, hdc, ticPos, flags & ~TBS_LEFT);
 }
 
-static COLORREF TRACKBAR_GetTicPenColor(const TRACKBAR_INFO *infoPtr)
-{
-#if __WINE_COMCTL32_VERSION == 6
+static void
+TRACKBAR_DrawTics (const TRACKBAR_INFO *infoPtr, HDC hdc)
+{
+    unsigned int i;
+    int ticFlags = infoPtr->dwStyle & 0x0f;
+    LOGPEN ticPen = { PS_SOLID, {1, 0}, GetSysColor (COLOR_3DDKSHADOW) };
+    HPEN hOldPen, hTicPen;
     HTHEME theme = GetWindowTheme (infoPtr->hwndSelf);
-
+    
     if (theme)
     {
         int part = (infoPtr->dwStyle & TBS_VERT) ? TKP_TICSVERT : TKP_TICS;
-        COLORREF color;
-
-        GetThemeColor (theme, part, TSS_NORMAL, TMT_COLOR, &color);
-        return color;
-    }
-#endif
-
-    return GetSysColor (COLOR_3DDKSHADOW);
-}
-
-static void
-TRACKBAR_DrawTics (const TRACKBAR_INFO *infoPtr, HDC hdc)
-{
-    unsigned int i;
-    int ticFlags = infoPtr->dwStyle & 0x0f;
-    LOGPEN ticPen;
-    HPEN hOldPen, hTicPen;
-
+        GetThemeColor (theme, part, TSS_NORMAL, TMT_COLOR, &ticPen.lopnColor);
+    }
     /* create the pen to draw the tics with */
-    ticPen.lopnStyle = PS_SOLID;
-    ticPen.lopnWidth.x = 1;
-    ticPen.lopnWidth.y = 0;
-    ticPen.lopnColor = TRACKBAR_GetTicPenColor(infoPtr);
     hTicPen = CreatePenIndirect(&ticPen);
     hOldPen = hTicPen ? SelectObject(hdc, hTicPen) : 0;
 
@@ -711,11 +696,9 @@
 static void
 TRACKBAR_DrawThumb (TRACKBAR_INFO *infoPtr, HDC hdc)
 {
+    HTHEME theme = GetWindowTheme (infoPtr->hwndSelf);
     int PointDepth;
     HBRUSH brush;
-
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme (infoPtr->hwndSelf);
 
     if (theme)
     {
@@ -743,7 +726,6 @@
         
         return;
     }
-#endif
 
     if (infoPtr->dwStyle & WS_DISABLED || infoPtr->flags & TB_DRAG_MODE)
     {
@@ -1255,8 +1237,6 @@
     {
         TRACKBAR_UpdateThumb(infoPtr);
         TRACKBAR_InvalidateThumbMove(infoPtr, oldPos, lPosition);
-
-        NotifyWinEvent( EVENT_OBJECT_VALUECHANGE, infoPtr->hwndSelf, OBJID_CLIENT, CHILDID_SELF );
     }
 
     return 0;
@@ -1579,8 +1559,8 @@
             SendMessageW (infoPtr->hwndToolTip, TTM_ADDTOOLW, 0, (LPARAM)&ti);
 	 }
     }
-
-    COMCTL32_OpenThemeForWindow (hwnd, L"Trackbar");
+    
+    OpenThemeData (hwnd, themeClass);
 
     return 0;
 }
@@ -1597,7 +1577,7 @@
     infoPtr->tics = NULL;
 
     SetWindowLongPtrW (infoPtr->hwndSelf, 0, 0);
-    COMCTL32_CloseThemeForWindow (infoPtr->hwndSelf);
+    CloseThemeData (GetWindowTheme (infoPtr->hwndSelf));
     Free (infoPtr);
 
     return 0;
@@ -1735,6 +1715,18 @@
     return 0;
 }
 
+
+/* update theme after a WM_THEMECHANGED message */
+static LRESULT theme_changed (const TRACKBAR_INFO* infoPtr)
+{
+    HTHEME theme = GetWindowTheme (infoPtr->hwndSelf);
+    CloseThemeData (theme);
+    OpenThemeData (infoPtr->hwndSelf, themeClass);
+    InvalidateRect (infoPtr->hwndSelf, NULL, FALSE);
+    return 0;
+}
+
+
 static LRESULT
 TRACKBAR_MouseMove (TRACKBAR_INFO *infoPtr, INT x, INT y)
 {
@@ -1753,7 +1745,7 @@
 
     if (!(infoPtr->flags & TB_DRAG_MODE)) 
     {
-        if (COMCTL32_IsThemed (infoPtr->hwndSelf))
+        if (GetWindowTheme (infoPtr->hwndSelf))
         {
             DWORD oldFlags = infoPtr->flags;
             POINT pt;
@@ -2023,7 +2015,7 @@
     case WM_GETOBJECT:
         if ((LONG)lParam == OBJID_QUERYCLASSNAMEIDX)
             return 0x10012;
-        return DefWindowProcW (hwnd, uMsg, wParam, lParam);
+        return 0;
 
     case WM_KEYDOWN:
         return TRACKBAR_KeyDown (infoPtr, (INT)wParam);
@@ -2062,7 +2054,7 @@
         return TRACKBAR_StyleChanged (infoPtr, wParam, (LPSTYLESTRUCT)lParam);
 
     case WM_THEMECHANGED:
-        return COMCTL32_ThemeChanged (infoPtr->hwndSelf, L"Trackbar", TRUE, FALSE);
+        return theme_changed (infoPtr);
 
     case WM_TIMER:
 	return TRACKBAR_Timer (infoPtr);
@@ -2093,3 +2085,9 @@
 
     RegisterClassW (&wndClass);
 }
+
+
+void TRACKBAR_Unregister (void)
+{
+    UnregisterClassW (TRACKBAR_CLASSW, NULL);
+}
