diff --git a/dlls/comctl32/datetime.c b/dlls/comctl32/datetime.c
--- a/dlls/comctl32/datetime.c
+++ b/dlls/comctl32/datetime.c
@@ -43,6 +43,8 @@
 #include "winnls.h"
 #include "commctrl.h"
 #include "comctl32.h"
+#include "uxtheme.h"
+#include "vsstyle.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(datetime);
@@ -125,6 +127,7 @@
 static BOOL DATETIME_SendDateTimeChangeNotify (const DATETIME_INFO *infoPtr);
 static const WCHAR allowedformatchars[] = L"dhHmMstyX";
 static const int maxrepetition [] = {4,2,2,2,4,2,2,4,-1};
+static const WCHAR *themeClass = WC_SCROLLBARW;
 
 /* valid date limits */
 static const SYSTEMTIME max_allowed_date = { .wYear = 9999, .wMonth = 12, .wDayOfWeek = 0, .wDay = 31 };
@@ -332,8 +335,6 @@
 
     DATETIME_UseFormat (infoPtr, format);
     InvalidateRect (infoPtr->hwndSelf, NULL, TRUE);
-
-    NotifyWinEvent(EVENT_OBJECT_VALUECHANGE, infoPtr->hwndSelf, OBJID_CLIENT, CHILDID_SELF);
 
     return TRUE;
 }
@@ -704,38 +705,12 @@
     return size.cx;
 }
 
-static void DATETIME_DrawBackground (DATETIME_INFO *infoPtr, HDC hdc)
-{
-#if __WINE_COMCTL32_VERSION == 6
+static void 
+DATETIME_Refresh (DATETIME_INFO *infoPtr, HDC hdc)
+{
     HTHEME theme;
-
-    theme = GetWindowTheme(infoPtr->hwndSelf);
-    if (theme)
-    {
-        int state;
-
-        if (infoPtr->dwStyle & WS_DISABLED)
-            state = ABS_DOWNDISABLED;
-        else if (infoPtr->bCalDepressed)
-            state = ABS_DOWNPRESSED;
-        else if (infoPtr->bCalHot)
-            state = ABS_DOWNHOT;
-        else
-            state = ABS_DOWNNORMAL;
-
-        DrawThemeBackground(theme, hdc, SBP_ARROWBTN, state, &infoPtr->calbutton, NULL);
-        return;
-    }
-#endif /* __WINE_COMCTL32_VERSION == 6 */
-
-    DrawFrameControl(hdc, &infoPtr->calbutton, DFC_SCROLL, DFCS_SCROLLDOWN |
-                     (infoPtr->bCalDepressed ? DFCS_PUSHED : 0) |
-                     (infoPtr->dwStyle & WS_DISABLED ? DFCS_INACTIVE : 0));
-}
-
-static void
-DATETIME_Refresh (DATETIME_INFO *infoPtr, HDC hdc)
-{
+    int state;
+
     TRACE("\n");
 
     if (infoPtr->dateValid) {
@@ -802,7 +777,26 @@
     if (infoPtr->dwStyle & DTS_UPDOWN)
         return;
 
-    DATETIME_DrawBackground(infoPtr, hdc);
+    theme = GetWindowTheme(infoPtr->hwndSelf);
+    if (theme)
+    {
+        if (infoPtr->dwStyle & WS_DISABLED)
+            state = ABS_DOWNDISABLED;
+        else if (infoPtr->bCalDepressed)
+            state = ABS_DOWNPRESSED;
+        else if (infoPtr->bCalHot)
+            state = ABS_DOWNHOT;
+        else
+            state = ABS_DOWNNORMAL;
+
+        DrawThemeBackground(theme, hdc, SBP_ARROWBTN, state, &infoPtr->calbutton, NULL);
+    }
+    else
+    {
+        DrawFrameControl(hdc, &infoPtr->calbutton, DFC_SCROLL,
+                         DFCS_SCROLLDOWN | (infoPtr->bCalDepressed ? DFCS_PUSHED : 0) |
+                         (infoPtr->dwStyle & WS_DISABLED ? DFCS_INACTIVE : 0) );
+    }
 }
 
 
@@ -1308,6 +1302,49 @@
     return 1;
 }
 
+static LRESULT DATETIME_NCPaint (HWND hwnd, HRGN region)
+{
+    INT cxEdge, cyEdge;
+    HRGN clipRgn;
+    HTHEME theme;
+    LONG exStyle;
+    RECT r;
+    HDC dc;
+
+    theme = OpenThemeDataForDpi(NULL, WC_EDITW, GetDpiForWindow(hwnd));
+    if (!theme)
+        return DefWindowProcW(hwnd, WM_NCPAINT, (WPARAM)region, 0);
+
+    exStyle = GetWindowLongW(hwnd, GWL_EXSTYLE);
+    if (!(exStyle & WS_EX_CLIENTEDGE))
+    {
+        CloseThemeData(theme);
+        return DefWindowProcW(hwnd, WM_NCPAINT, (WPARAM)region, 0);
+    }
+
+    cxEdge = GetSystemMetrics(SM_CXEDGE);
+    cyEdge = GetSystemMetrics(SM_CYEDGE);
+    GetWindowRect(hwnd, &r);
+
+    /* New clipping region passed to default proc to exclude border */
+    clipRgn = CreateRectRgn(r.left + cxEdge, r.top + cyEdge, r.right - cxEdge, r.bottom - cyEdge);
+    if (region != (HRGN)1)
+        CombineRgn(clipRgn, clipRgn, region, RGN_AND);
+    OffsetRect(&r, -r.left, -r.top);
+
+    dc = GetDCEx(hwnd, region, DCX_WINDOW | DCX_INTERSECTRGN);
+    if (IsThemeBackgroundPartiallyTransparent(theme, 0, 0))
+        DrawThemeParentBackground(hwnd, dc, &r);
+    DrawThemeBackground(theme, dc, 0, 0, &r, 0);
+    ReleaseDC(hwnd, dc);
+    CloseThemeData(theme);
+
+    /* Call default proc to get the scrollbars etc. also painted */
+    DefWindowProcW(hwnd, WM_NCPAINT, (WPARAM)clipRgn, 0);
+    DeleteObject(clipRgn);
+    return 0;
+}
+
 static LRESULT DATETIME_MouseMove (DATETIME_INFO *infoPtr, LONG x, LONG y)
 {
     TRACKMOUSEEVENT event;
@@ -1490,6 +1527,17 @@
     return 0;
 }
 
+static LRESULT DATETIME_ThemeChanged (DATETIME_INFO *infoPtr)
+{
+    HTHEME theme;
+
+    theme = GetWindowTheme(infoPtr->hwndSelf);
+    CloseThemeData(theme);
+    OpenThemeData(infoPtr->hwndSelf, themeClass);
+    InvalidateRect(infoPtr->hwndSelf, NULL, TRUE);
+    return 0;
+}
+
 static BOOL DATETIME_GetIdealSize(DATETIME_INFO *infoPtr, SIZE *size)
 {
     SIZE field_size;
@@ -1583,7 +1631,7 @@
     infoPtr->hFont = GetStockObject(DEFAULT_GUI_FONT);
 
     SetWindowLongPtrW (hwnd, 0, (DWORD_PTR)infoPtr);
-    COMCTL32_OpenThemeForWindow(hwnd, WC_SCROLLBARW);
+    OpenThemeData(hwnd, themeClass);
 
     return 0;
 }
@@ -1593,7 +1641,10 @@
 static LRESULT
 DATETIME_Destroy (DATETIME_INFO *infoPtr)
 {
-    COMCTL32_CloseThemeForWindow(infoPtr->hwndSelf);
+    HTHEME theme;
+
+    theme = GetWindowTheme(infoPtr->hwndSelf);
+    CloseThemeData(theme);
 
     if (infoPtr->hwndCheckbut)
 	DestroyWindow(infoPtr->hwndCheckbut);
@@ -1721,7 +1772,7 @@
         return DATETIME_NCCreate (hwnd, (LPCREATESTRUCTW)lParam);
 
     case WM_NCPAINT:
-        return COMCTL32_NCPaint(hwnd, wParam, lParam, WC_EDITW);
+        return DATETIME_NCPaint(hwnd, (HRGN)wParam);
 
     case WM_MOUSEMOVE:
         return DATETIME_MouseMove(infoPtr, (SHORT)LOWORD(lParam), (SHORT)HIWORD(lParam));
@@ -1760,7 +1811,7 @@
         return DATETIME_StyleChanged(infoPtr, wParam, (LPSTYLESTRUCT)lParam);
 
     case WM_THEMECHANGED:
-        return COMCTL32_ThemeChanged(infoPtr->hwndSelf, WC_SCROLLBARW, TRUE, TRUE);
+        return DATETIME_ThemeChanged(infoPtr);
 
     case WM_SETFONT:
         return DATETIME_SetFont(infoPtr, (HFONT)wParam, (BOOL)lParam);
@@ -1801,3 +1852,10 @@
 
     RegisterClassW (&wndClass);
 }
+
+
+void
+DATETIME_Unregister (void)
+{
+    UnregisterClassW (DATETIMEPICK_CLASSW, NULL);
+}
