diff --git a/dlls/winex11.drv/clipboard.c b/dlls/winex11.drv/clipboard.c
--- a/dlls/winex11.drv/clipboard.c
+++ b/dlls/winex11.drv/clipboard.c
@@ -83,6 +83,7 @@
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
 #include "x11drv.h"
+#include "xfixes.h"
 
 #ifdef HAVE_X11_EXTENSIONS_XFIXES_H
 #include <X11/extensions/Xfixes.h>
@@ -199,7 +200,6 @@
 static ULONG last_clipboard_update;
 static struct clipboard_format **current_x11_formats;
 static unsigned int nb_current_x11_formats;
-static BOOL use_xfixes;
 
 Display *clipboard_display = NULL;
 
@@ -580,6 +580,101 @@
 }
 
 
+/* based on wine_get_dos_file_name */
+static WCHAR *get_dos_file_name( const char *path )
+{
+    ULONG len = strlen( path ) + 9; /* \??\unix prefix */
+    WCHAR *ret;
+
+    if (!(ret = malloc( len * sizeof(WCHAR) ))) return NULL;
+    if (wine_unix_to_nt_file_name( path, ret, &len ))
+    {
+        free( ret );
+        return NULL;
+    }
+
+    if (ret[5] == ':')
+    {
+        /* get rid of the \??\ prefix */
+        memmove( ret, ret + 4, (len - 4) * sizeof(WCHAR) );
+    }
+    else ret[1] = '\\';
+    return ret;
+}
+
+
+/***********************************************************************
+ *           get_nt_pathname
+ *
+ * Simplified version of RtlDosPathNameToNtPathName_U.
+ */
+static BOOL get_nt_pathname( const WCHAR *name, UNICODE_STRING *nt_name )
+{
+    static const WCHAR ntprefixW[] = {'\\','?','?','\\'};
+    static const WCHAR uncprefixW[] = {'U','N','C','\\'};
+    size_t len = lstrlenW( name );
+    WCHAR *ptr;
+
+    nt_name->MaximumLength = (len + 8) * sizeof(WCHAR);
+    if (!(ptr = malloc( nt_name->MaximumLength ))) return FALSE;
+    nt_name->Buffer = ptr;
+
+    memcpy( ptr, ntprefixW, sizeof(ntprefixW) );
+    ptr += ARRAYSIZE(ntprefixW);
+    if (name[0] == '\\' && name[1] == '\\')
+    {
+        if ((name[2] == '.' || name[2] == '?') && name[3] == '\\')
+        {
+            name += 4;
+            len -= 4;
+        }
+        else
+        {
+            memcpy( ptr, uncprefixW, sizeof(uncprefixW) );
+            ptr += ARRAYSIZE(uncprefixW);
+            name += 2;
+            len -= 2;
+        }
+    }
+    memcpy( ptr, name, (len + 1) * sizeof(WCHAR) );
+    ptr += len;
+    nt_name->Length = (ptr - nt_name->Buffer) * sizeof(WCHAR);
+    return TRUE;
+}
+
+
+/* based on wine_get_unix_file_name */
+static char *get_unix_file_name( const WCHAR *dosW )
+{
+    UNICODE_STRING nt_name;
+    OBJECT_ATTRIBUTES attr;
+    NTSTATUS status;
+    ULONG size = 256;
+    char *buffer;
+
+    if (!get_nt_pathname( dosW, &nt_name )) return NULL;
+    InitializeObjectAttributes( &attr, &nt_name, 0, 0, NULL );
+    for (;;)
+    {
+        if (!(buffer = malloc( size )))
+        {
+            free( nt_name.Buffer );
+            return NULL;
+        }
+        status = wine_nt_to_unix_file_name( &attr, buffer, &size, FILE_OPEN_IF );
+        if (status != STATUS_BUFFER_TOO_SMALL) break;
+        free( buffer );
+    }
+    free( nt_name.Buffer );
+    if (status)
+    {
+        free( buffer );
+        return NULL;
+    }
+    return buffer;
+}
+
+
 static CPTABLEINFO *get_xstring_cp(void)
 {
     static CPTABLEINFO cp;
@@ -656,7 +751,7 @@
             if (uri[7] == '/')
             {
                 /* file:///path/to/file (nautilus, thunar) */
-                ntdll_get_dos_file_name( &uri[7], &ret, FILE_OPEN );
+                ret = get_dos_file_name( &uri[7] );
             }
             else if (uri[7])
             {
@@ -666,11 +761,18 @@
                 if (path)
                 {
                     *path = '\0';
-                    if (!strcmp(&uri[7], "localhost") ||
-                        (!gethostname(hostname, sizeof(hostname)) && !strcmp(hostname, &uri[7])))
+                    if (strcmp(&uri[7], "localhost") == 0)
                     {
                         *path = '/';
-                        ntdll_get_dos_file_name( path, &ret, FILE_OPEN );
+                        ret = get_dos_file_name( path );
+                    }
+                    else if (gethostname(hostname, sizeof(hostname)) == 0)
+                    {
+                        if (strcmp(hostname, &uri[7]) == 0)
+                        {
+                            *path = '/';
+                            ret = get_dos_file_name( path );
+                        }
                     }
                 }
             }
@@ -678,7 +780,7 @@
         else if (uri[6])
         {
             /* file:/path/to/file (konqueror) */
-            ntdll_get_dos_file_name( &uri[5], &ret, FILE_OPEN );
+            ret = get_dos_file_name( &uri[5] );
         }
     }
     free( uri );
@@ -955,7 +1057,7 @@
 
     for (ptr = data; ptr < (const char *)data + size; ptr += strlen( ptr ) + 1)
     {
-        ntdll_get_dos_file_name( ptr, &path, FILE_OPEN );
+        path = get_dos_file_name( ptr );
 
         TRACE( "converted URI %s to DOS path %s\n", debugstr_a(ptr), debugstr_w(path) );
 
@@ -1490,9 +1592,9 @@
         char *unixFilename = NULL;
         UINT uriSize;
         UINT u;
-        NTSTATUS status = ntdll_get_unix_file_name( ptr, &unixFilename, FILE_OPEN );
-
-        if (status) goto failed;
+
+        unixFilename = get_unix_file_name( ptr );
+        if (unixFilename == NULL) goto failed;
         ptr += lstrlenW( ptr ) + 1;
 
         uriSize = 8 + /* file:/// */
@@ -2068,28 +2170,6 @@
 static void xfixes_init(void)
 {
 #ifdef SONAME_LIBXFIXES
-    typeof(XFixesSelectSelectionInput) *pXFixesSelectSelectionInput;
-    typeof(XFixesQueryExtension) *pXFixesQueryExtension;
-    typeof(XFixesQueryVersion) *pXFixesQueryVersion;
-
-    int event_base, error_base;
-    int major = 3, minor = 0;
-    void *handle;
-
-    handle = dlopen(SONAME_LIBXFIXES, RTLD_NOW);
-    if (!handle) return;
-
-    pXFixesQueryExtension = dlsym(handle, "XFixesQueryExtension");
-    if (!pXFixesQueryExtension) return;
-    pXFixesQueryVersion = dlsym(handle, "XFixesQueryVersion");
-    if (!pXFixesQueryVersion) return;
-    pXFixesSelectSelectionInput = dlsym(handle, "XFixesSelectSelectionInput");
-    if (!pXFixesSelectSelectionInput) return;
-
-    if (!pXFixesQueryExtension(clipboard_display, &event_base, &error_base))
-        return;
-    pXFixesQueryVersion(clipboard_display, &major, &minor);
-    use_xfixes = (major >= 1);
     if (!use_xfixes) return;
 
     pXFixesSelectSelectionInput(clipboard_display, import_window, x11drv_atom(CLIPBOARD),
@@ -2103,7 +2183,7 @@
                 XFixesSelectionWindowDestroyNotifyMask |
                 XFixesSelectionClientCloseNotifyMask);
     }
-    X11DRV_register_event_handler(event_base + XFixesSelectionNotify,
+    X11DRV_register_event_handler(xfixes_event_base + XFixesSelectionNotify,
             selection_notify_event, "XFixesSelectionNotify");
     TRACE("xfixes succesully initialized\n");
 #else
