diff --git a/dlls/comctl32/tests/updown.c b/dlls/comctl32/tests/updown.c
--- a/dlls/comctl32/tests/updown.c
+++ b/dlls/comctl32/tests/updown.c
@@ -48,7 +48,6 @@
 #include <stdio.h>
 
 #include "wine/test.h"
-#include "v6util.h"
 #include "msg.h"
 
 #define expect(expected,got) expect_(__LINE__, expected, got)
@@ -72,47 +71,6 @@
 static BOOL (WINAPI *pSetWindowSubclass)(HWND, SUBCLASSPROC, UINT_PTR, DWORD_PTR);
 
 static struct msg_sequence *sequences[NUM_MSG_SEQUENCES];
-
-static void CALLBACK msg_winevent_proc(HWINEVENTHOOK hevent,
-                                       DWORD event,
-                                       HWND hwnd,
-                                       LONG object_id,
-                                       LONG child_id,
-                                       DWORD thread_id,
-                                       DWORD event_time)
-{
-    struct message msg = {0};
-    char class_name[256];
-
-    /* ignore window and other system events */
-    if (object_id != OBJID_CLIENT) return;
-
-    /* ignore events not from an updown control */
-    if (!GetClassNameA(hwnd, class_name, ARRAY_SIZE(class_name)) ||
-        strcmp(class_name, UPDOWN_CLASSA) != 0)
-        return;
-
-    msg.message = event;
-    msg.flags = winevent_hook|wparam|lparam;
-    msg.wParam = object_id;
-    msg.lParam = child_id;
-    add_message(sequences, UPDOWN_SEQ_INDEX, &msg);
-}
-
-static void init_winevent_hook(void) {
-    hwineventhook = SetWinEventHook(EVENT_MIN, EVENT_MAX, GetModuleHandleA(0), msg_winevent_proc,
-                                    0, GetCurrentThreadId(), WINEVENT_INCONTEXT);
-    if (!hwineventhook)
-        win_skip( "no win event hook support\n" );
-}
-
-static void uninit_winevent_hook(void) {
-    if (!hwineventhook)
-        return;
-
-    UnhookWinEvent(hwineventhook);
-    hwineventhook = 0;
-}
 
 static const struct message add_updown_with_edit_seq[] = {
     { WM_WINDOWPOSCHANGING, sent },
@@ -138,19 +96,14 @@
     { UDM_SETRANGE, sent|lparam, 0, MAKELONG(100,0) },
     { UDM_GETRANGE, sent},
     { UDM_SETPOS, sent|lparam, 0, 5},
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS, sent},
     { UDM_SETPOS, sent|lparam, 0, 0},
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS, sent},
     { UDM_SETPOS, sent|lparam, 0, MAKELONG(-1,0)},
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS, sent},
     { UDM_SETPOS, sent|lparam, 0, 100},
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS, sent},
     { UDM_SETPOS, sent|lparam, 0, 101},
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS, sent},
     { 0 }
 };
@@ -159,19 +112,14 @@
     { UDM_SETRANGE32, sent|lparam, 0, 1000 },
     { UDM_GETRANGE32, sent}, /* Cannot check wparam and lparam as they are ptrs */
     { UDM_SETPOS32, sent|lparam, 0, 500 },
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS32, sent},
     { UDM_SETPOS32, sent|lparam, 0, 0 },
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS32, sent},
     { UDM_SETPOS32, sent|lparam, 0, -1 },
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS32, sent},
     { UDM_SETPOS32, sent|lparam, 0, 1000 },
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS32, sent},
     { UDM_SETPOS32, sent|lparam, 0, 1001 },
-    { EVENT_OBJECT_VALUECHANGE, winevent_hook|wparam|lparam, OBJID_CLIENT, 0 },
     { UDM_GETPOS32, sent},
     { 0 }
 };
@@ -604,7 +552,7 @@
     DestroyWindow(updown);
 }
 
-static void test_updown_buddy(BOOL is_v6)
+static void test_updown_buddy(void)
 {
     HWND updown, buddyReturn, buddy;
     RECT rect, rect2;
@@ -647,7 +595,7 @@
     updown= create_updown_control(UDS_ALIGNRIGHT | UDS_ARROWKEYS, buddy);
     ok(proc != (WNDPROC)GetWindowLongPtrA(buddy, GWLP_WNDPROC), "Subclassing expected\n");
 
-    if (pSetWindowSubclass && ! is_v6)
+    if (pSetWindowSubclass)
     {
         /* updown uses subclass helpers for buddy on >5.8x systems */
         ok(GetPropA(buddy, "CC32SubclassInfo") != NULL, "Expected CC32SubclassInfo property\n");
@@ -1065,46 +1013,6 @@
     DestroyWindow(updown);
 }
 
-static void test_accel(void)
-{
-    UDACCEL accel[4];
-    HWND updown;
-    int r;
-
-    updown = create_updown_control(UDS_ALIGNRIGHT, g_edit);
-
-    flush_sequences(sequences, NUM_MSG_SEQUENCES);
-
-    r = SendMessageA(updown, UDM_GETACCEL, 0, (LPARAM)accel);
-    expect(3, r);
-
-    r = SendMessageA(updown, UDM_GETACCEL, 1, (LPARAM)accel);
-    expect(3, r);
-    expect(0, accel[0].nSec);
-    expect(1, accel[0].nInc);
-
-    r = SendMessageA(updown, UDM_GETACCEL, 4, (LPARAM)accel);
-    expect(3, r);
-    expect(0, accel[0].nSec);
-    expect(1, accel[0].nInc);
-    expect(2, accel[1].nSec);
-    expect(5, accel[1].nInc);
-    expect(5, accel[2].nSec);
-    expect(20, accel[2].nInc);
-
-    accel[0].nSec = 0;
-    accel[0].nInc = 5;
-    r = SendMessageA(updown, UDM_SETACCEL, 1, (LPARAM)accel);
-    expect(TRUE, r);
-
-    r = SendMessageA(updown, UDM_GETACCEL, 3, (LPARAM)accel);
-    expect(1, r);
-    expect(0, accel[0].nSec);
-    expect(5, accel[0].nInc);
-
-    DestroyWindow(updown);
-}
-
 static void init_functions(void)
 {
     HMODULE hComCtl32 = LoadLibraryA("comctl32.dll");
@@ -1119,9 +1027,6 @@
 
 START_TEST(updown)
 {
-    ULONG_PTR ctx_cookie;
-    HANDLE hCtx;
-
     init_functions();
 
     init_msg_sequences(sequences, NUM_MSG_SEQUENCES);
@@ -1131,49 +1036,16 @@
     g_edit = create_edit_control();
     ok(g_edit != NULL, "Failed to create edit control\n");
 
-    init_winevent_hook();
-
     test_updown_create();
     test_updown_pos();
     test_updown_pos32();
-    test_updown_buddy(FALSE);
+    test_updown_buddy();
     test_updown_base();
     test_updown_unicode();
     test_UDS_SETBUDDY();
     test_UDS_SETBUDDYINT();
     test_CreateUpDownControl();
     test_updown_pos_notifications();
-    test_accel();
-
-    DestroyWindow(g_edit);
-
-    if (!load_v6_module(&ctx_cookie, &hCtx))
-    {
-        DestroyWindow(parent_wnd);
-        return;
-    }
-
-    init_functions();
-
-    g_edit = create_edit_control();
-    ok(g_edit != NULL, "Failed to create edit control\n");
-
-    /* comctl32 version 6 tests start here */
-    test_updown_create();
-    test_updown_pos();
-    test_updown_pos32();
-    test_updown_buddy(TRUE);
-    test_updown_base();
-    test_updown_unicode();
-    test_UDS_SETBUDDY();
-    test_UDS_SETBUDDYINT();
-    test_CreateUpDownControl();
-    test_updown_pos_notifications();
-    test_accel();
-
-    uninit_winevent_hook();
-
-    unload_v6_module(ctx_cookie, hCtx);
 
     DestroyWindow(g_edit);
     DestroyWindow(parent_wnd);
