diff --git a/dlls/winemac.drv/cocoa_app.m b/dlls/winemac.drv/cocoa_app.m
--- a/dlls/winemac.drv/cocoa_app.m
+++ b/dlls/winemac.drv/cocoa_app.m
@@ -18,6 +18,8 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
+#import <Carbon/Carbon.h>
+
 #import "cocoa_app.h"
 #import "cocoa_cursorclipping.h"
 #import "cocoa_event.h"
@@ -41,7 +43,16 @@
 static NSString* const WineActivatingAppConfigDirKey = @"ActivatingAppConfigDir";
 
 
-bool macdrv_err_on;
+int macdrv_err_on;
+
+
+#if !defined(MAC_OS_X_VERSION_10_12) || MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_X_VERSION_10_12
+@interface NSWindow (WineAutoTabbingExtensions)
+
+    + (void) setAllowsAutomaticWindowTabbing:(BOOL)allows;
+
+@end
+#endif
 
 
 #if !defined(MAC_OS_VERSION_14_0) || MAC_OS_X_VERSION_MAX_ALLOWED < MAC_OS_VERSION_14_0
@@ -134,7 +145,8 @@
 
             [[NSUserDefaults standardUserDefaults] registerDefaults:defaults];
 
-            [NSWindow setAllowsAutomaticWindowTabbing:NO];
+            if ([NSWindow respondsToSelector:@selector(setAllowsAutomaticWindowTabbing:)])
+                [NSWindow setAllowsAutomaticWindowTabbing:NO];
         }
     }
 
@@ -178,6 +190,15 @@
             latentDisplayModes = [[NSMutableDictionary alloc] init];
 
             windowsBeingDragged = [[NSMutableSet alloc] init];
+
+            // On macOS 10.12+, use notifications to more reliably detect when windows are being dragged.
+            if ([NSProcessInfo instancesRespondToSelector:@selector(isOperatingSystemAtLeastVersion:)])
+            {
+                NSOperatingSystemVersion requiredVersion = { 10, 12, 0 };
+                useDragNotifications = [[NSProcessInfo processInfo] isOperatingSystemAtLeastVersion:requiredVersion];
+            }
+            else
+                useDragNotifications = NO;
 
             if (!requests || !requestsManipQueue || !eventQueues || !eventQueuesLock ||
                 !keyWindows || !originalDisplayModes || !latentDisplayModes)
@@ -235,11 +256,13 @@
             if (activateIfTransformed)
                 [self tryToActivateIgnoringOtherApps:YES];
 
-            if (!enable_app_nap)
+#if defined(MAC_OS_X_VERSION_10_9) && MAC_OS_X_VERSION_MAX_ALLOWED >= MAC_OS_X_VERSION_10_9
+            if (!enable_app_nap && [NSProcessInfo instancesRespondToSelector:@selector(beginActivityWithOptions:reason:)])
             {
                 [[[NSProcessInfo processInfo] beginActivityWithOptions:NSActivityUserInitiatedAllowingIdleSystemSleep
                                                                 reason:@"Running Windows program"] retain]; // intentional leak
             }
+#endif
 
             mainMenu = [[[NSMenu alloc] init] autorelease];
 
@@ -305,7 +328,7 @@
         }
     }
 
-    - (BOOL) waitUntilQueryDone:(bool*)done timeout:(NSDate*)timeout processEvents:(BOOL)processEvents
+    - (BOOL) waitUntilQueryDone:(int*)done timeout:(NSDate*)timeout processEvents:(BOOL)processEvents
     {
         PerformRequest(NULL);
 
@@ -579,8 +602,7 @@
         NSUInteger nextFloatingIndex = 0;
         __block NSInteger maxLevel = NSIntegerMin;
         __block NSInteger maxNonfloatingLevel = NSNormalWindowLevel;
-        /* Windows with WS_EX_TOPMOST should have a window level higher than the macOS dock */
-        __block NSInteger minFloatingLevel = kCGDockWindowLevel + 1;
+        __block NSInteger minFloatingLevel = NSFloatingWindowLevel;
         __block WineWindow* prev = nil;
         WineWindow* window;
 
@@ -1824,6 +1846,17 @@
                     [window postKeyEvent:anEvent];
             }
         }
+        else if (!useDragNotifications && type == NSEventTypeAppKitDefined)
+        {
+            WineWindow *window = (WineWindow *)[anEvent window];
+            short subtype = [anEvent subtype];
+
+            // These subtypes are not documented but they appear to mean
+            // "a window is being dragged" and "a window is no longer being
+            // dragged", respectively.
+            if ((subtype == 20 || subtype == 21) && [window isKindOfClass:[WineWindow class]])
+                [self handleWindowDrag:window begin:(subtype == 20)];
+        }
 
         return ret;
     }
@@ -1883,23 +1916,25 @@
             [windowsBeingDragged removeObject:window];
         }];
 
-        [nc addObserverForName:NSWindowWillStartDraggingNotification
-                        object:nil
-                         queue:[NSOperationQueue mainQueue]
-                    usingBlock:^(NSNotification *note){
-            NSWindow* window = [note object];
-            if ([window isKindOfClass:[WineWindow class]])
-                [self handleWindowDrag:(WineWindow *)window begin:YES];
-        }];
-
-        [nc addObserverForName:NSWindowDidEndDraggingNotification
-                        object:nil
-                         queue:[NSOperationQueue mainQueue]
-                    usingBlock:^(NSNotification *note){
-            NSWindow* window = [note object];
-            if ([window isKindOfClass:[WineWindow class]])
-                [self handleWindowDrag:(WineWindow *)window begin:NO];
-        }];
+        if (useDragNotifications) {
+            [nc addObserverForName:NSWindowWillStartDraggingNotification
+                            object:nil
+                             queue:[NSOperationQueue mainQueue]
+                        usingBlock:^(NSNotification *note){
+                NSWindow* window = [note object];
+                if ([window isKindOfClass:[WineWindow class]])
+                    [self handleWindowDrag:(WineWindow *)window begin:YES];
+            }];
+
+            [nc addObserverForName:NSWindowDidEndDraggingNotification
+                            object:nil
+                             queue:[NSOperationQueue mainQueue]
+                        usingBlock:^(NSNotification *note){
+                NSWindow* window = [note object];
+                if ([window isKindOfClass:[WineWindow class]])
+                    [self handleWindowDrag:(WineWindow *)window begin:NO];
+            }];
+        }
 
         [nc addObserver:self
                selector:@selector(keyboardSelectionDidChange)
@@ -2137,7 +2172,7 @@
         [bestOption deminiaturize:self];
     }
 
-    - (void) setRetinaMode:(BOOL)mode
+    - (void) setRetinaMode:(int)mode
     {
         retina_on = mode;
 
@@ -2396,7 +2431,7 @@
  *
  * Returns the keyboard layout uchr data, keyboard type and input source.
  */
-void macdrv_get_input_source_info(CFDataRef* uchr, CGEventSourceKeyboardType* keyboard_type, bool* is_iso, TISInputSourceRef* input_source)
+void macdrv_get_input_source_info(CFDataRef* uchr, CGEventSourceKeyboardType* keyboard_type, int* is_iso, TISInputSourceRef* input_source)
 {
     OnMainThread(^{
         TISInputSourceRef inputSourceLayout;
@@ -2432,12 +2467,13 @@
 /***********************************************************************
  *              macdrv_set_display_mode
  */
-int macdrv_set_display_mode(CGDirectDisplayID displayID, CGDisplayModeRef display_mode)
+int macdrv_set_display_mode(const struct macdrv_display* display,
+                            CGDisplayModeRef display_mode)
 {
     __block int ret;
 
     OnMainThread(^{
-        ret = [[WineApplicationController sharedController] setMode:display_mode forDisplay:displayID];
+        ret = [[WineApplicationController sharedController] setMode:display_mode forDisplay:display->displayID];
     });
 
     return ret;
@@ -2609,9 +2645,9 @@
 /***********************************************************************
  *              macdrv_using_input_method
  */
-bool macdrv_using_input_method(void)
+int macdrv_using_input_method(void)
 {
-    __block bool ret;
+    __block BOOL ret;
 
     OnMainThread(^{
         ret = [[WineApplicationController sharedController] inputSourceIsInputMethod];
@@ -2683,9 +2719,9 @@
     return ret;
 }
 
-bool macdrv_select_input_source(TISInputSourceRef input_source)
+int macdrv_select_input_source(TISInputSourceRef input_source)
 {
-    __block bool ret = false;
+    __block int ret = FALSE;
 
     OnMainThread(^{
         ret = (TISSelectInputSource(input_source) == noErr);
@@ -2694,16 +2730,16 @@
     return ret;
 }
 
-void macdrv_set_cocoa_retina_mode(bool new_mode)
+void macdrv_set_cocoa_retina_mode(int new_mode)
 {
     OnMainThread(^{
         [[WineApplicationController sharedController] setRetinaMode:new_mode];
     });
 }
 
-bool macdrv_is_any_wine_window_visible(void)
+int macdrv_is_any_wine_window_visible(void)
 {
-    __block bool ret = false;
+    __block int ret = FALSE;
 
     OnMainThread(^{
         ret = [[WineApplicationController sharedController] isAnyWineWindowVisible];
