diff --git a/dlls/crypt32/tests/store.c b/dlls/crypt32/tests/store.c
--- a/dlls/crypt32/tests/store.c
+++ b/dlls/crypt32/tests/store.c
@@ -117,19 +117,6 @@
 0x8e,0xe7,0x5f,0x76,0x2b,0xd1,0x6a,0x82,0xb3,0x30,0x25,0x61,0xf6,0x25,0x23,
 0x57,0x6c,0x0b,0x47,0xb8 };
 
-static const BYTE certWithUsage[] = { 0x30, 0x81, 0x93, 0x02, 0x01, 0x01, 0x30,
- 0x02, 0x06, 0x00, 0x30, 0x15, 0x31, 0x13, 0x30, 0x11, 0x06, 0x03, 0x55, 0x04,
- 0x03, 0x13, 0x0a, 0x4a, 0x75, 0x61, 0x6e, 0x20, 0x4c, 0x61, 0x6e, 0x67, 0x00,
- 0x30, 0x22, 0x18, 0x0f, 0x31, 0x36, 0x30, 0x31, 0x30, 0x31, 0x30, 0x31, 0x30,
- 0x30, 0x30, 0x30, 0x30, 0x30, 0x5a, 0x18, 0x0f, 0x31, 0x36, 0x30, 0x31, 0x30,
- 0x31, 0x30, 0x31, 0x30, 0x30, 0x30, 0x30, 0x30, 0x30, 0x5a, 0x30, 0x15, 0x31,
- 0x13, 0x30, 0x11, 0x06, 0x03, 0x55, 0x04, 0x03, 0x13, 0x0a, 0x4a, 0x75, 0x61,
- 0x6e, 0x20, 0x4c, 0x61, 0x6e, 0x67, 0x00, 0x30, 0x07, 0x30, 0x02, 0x06, 0x00,
- 0x03, 0x01, 0x00, 0xa3, 0x2f, 0x30, 0x2d, 0x30, 0x2b, 0x06, 0x03, 0x55, 0x1d,
- 0x25, 0x01, 0x01, 0xff, 0x04, 0x21, 0x30, 0x1f, 0x06, 0x08, 0x2b, 0x06, 0x01,
- 0x05, 0x05, 0x07, 0x03, 0x03, 0x06, 0x08, 0x2b, 0x06, 0x01, 0x05, 0x05, 0x07,
- 0x03, 0x02, 0x06, 0x09, 0x2a, 0x86, 0x48, 0x86, 0xf7, 0x0d, 0x01, 0x01, 0x01 };
-
 #define test_store_is_empty(store) _test_store_is_empty(__LINE__,store)
 static void _test_store_is_empty(unsigned line, HCERTSTORE store)
 {
@@ -141,16 +128,10 @@
 
 static void testMemStore(void)
 {
-    static const BYTE cert1_hash[] =
-        { 0x4a, 0x7f, 0x32, 0x1f, 0xcf, 0x3b, 0xc0, 0x87, 0x48, 0x2b, 0xa1, 0x86, 0x54, 0x18, 0xe4, 0x3a, 0x0e, 0x53, 0x7e, 0x2b, };
-    static const BYTE cert2_hash[] =
-        { 0xef, 0x7c, 0x6f, 0x27, 0x13, 0xff, 0x84, 0x14, 0x7c, 0xb5, 0xf0, 0x0c, 0x5a, 0x82, 0x91, 0x78, 0x96, 0x82, 0x19, 0x5c, };
     HCERTSTORE store1, store2;
-    PCCERT_CONTEXT context, stored[2], copy;
-    unsigned int count;
-    BYTE hash[20];
+    PCCERT_CONTEXT context;
     BOOL ret;
-    DWORD GLE, size;
+    DWORD GLE;
 
     /* NULL provider */
     store1 = CertOpenStore(0, 0, 0, 0, NULL);
@@ -285,71 +266,13 @@
         ok(!context, "Expected an empty store\n");
     }
 
-    /* Test deleting and adding cert during enumeration. */
-    ret = CertAddEncodedCertificateToStore(store1, X509_ASN_ENCODING, bigCert,
-     sizeof(bigCert), CERT_STORE_ADD_ALWAYS, NULL);
-    ok(ret, "CertAddEncodedCertificateToStore failed: %08lx\n", GetLastError());
-    ret = CertAddEncodedCertificateToStore(store1, X509_ASN_ENCODING, bigCert2,
-     sizeof(bigCert2), CERT_STORE_ADD_ALWAYS, NULL);
-    ok(ret, "CertAddEncodedCertificateToStore failed: %08lx\n", GetLastError());
-
-    context = NULL;
-    count = 0;
-    while ((context = CertEnumCertificatesInStore(store1, context)))
-    {
-        ok(count < 2, "got %u.\n", count);
-        stored[count] = CertDuplicateCertificateContext(context);
-        ++count;
-    }
-    ok(count == 2, "got %u.\n", count);
-
-    count = 0;
-    context = NULL;
-    while ((context = CertEnumCertificatesInStore(store1, context)))
-    {
-        size = sizeof(hash);
-        ret = CertGetCertificateContextProperty(context, CERT_HASH_PROP_ID, hash, &size);
-        ok(ret, "CertGetCertificateContextProperty failed: %08lx\n", GetLastError());
-        ok(!memcmp(hash, cert1_hash, sizeof(hash)), "wrong cert.\n");
-        copy = CertDuplicateCertificateContext(context);
-        ok(copy == context, "got %p, %p.\n", copy, context);
-        ret = CertDeleteCertificateFromStore(copy);
-        ok(ret, "CertDeleteCertificateFromStore failed: %08lx\n", GetLastError());
-        if (!count)
-        {
-            ret = CertDeleteCertificateFromStore(stored[1]);
-            ok(ret, "CertDeleteCertificateFromStore failed: %08lx\n", GetLastError());
-            ret = CertAddEncodedCertificateToStore(store1, X509_ASN_ENCODING, certWithUsage,
-             sizeof(certWithUsage), CERT_STORE_ADD_ALWAYS, NULL);
-            ok(ret, "CertAddEncodedCertificateToStore failed: %08lx\n", GetLastError());
-        }
-        ++count;
-    }
-    ok(count == 1, "got %u.\n", count);
-
-    count = 0;
-    context = NULL;
-    while ((context = CertEnumCertificatesInStore(store1, context)))
-    {
-        size = sizeof(hash);
-        ret = CertGetCertificateContextProperty(context, CERT_HASH_PROP_ID, hash, &size);
-        ok(ret, "CertGetCertificateContextProperty failed: %08lx\n", GetLastError());
-        ok(!memcmp(hash, cert2_hash, sizeof(hash)), "wrong cert.\n");
-        copy = CertDuplicateCertificateContext(context);
-        ret = CertDeleteCertificateFromStore(copy);
-        ok(ret, "CertDeleteCertificateFromStore failed: %08lx\n", GetLastError());
-        ++count;
-    }
-    ok(count == 1, "got %u.\n", count);
-    CertFreeCertificateContext(stored[0]);
-
     /* close an empty store */
     ret = CertCloseStore(NULL, 0);
     ok(ret, "CertCloseStore failed: %ld\n", GetLastError());
-    ret = CertCloseStore(store1, CERT_CLOSE_STORE_CHECK_FLAG);
-    ok(ret, "got error %#lx.\n", GetLastError());
-    ret = CertCloseStore(store2, CERT_CLOSE_STORE_CHECK_FLAG);
-    ok(ret, "got error %#lx.\n", GetLastError());
+    ret = CertCloseStore(store1, 0);
+    ok(ret, "CertCloseStore failed: %ld\n", GetLastError());
+    ret = CertCloseStore(store2, 0);
+    ok(ret, "CertCloseStore failed: %ld\n", GetLastError());
 
     /* This seems nonsensical, but you can open a read-only mem store, only
      * it isn't read-only
@@ -524,8 +447,8 @@
         ok (ret, "Failed to delete certificate from store at %ld, %lx\n", i, GetLastError());
 
         CertFreeCertificateContext(cert1);
-        ret = CertCloseStore(store, CERT_CLOSE_STORE_CHECK_FLAG);
-        ok(ret, "got error %#lx.\n", GetLastError());
+        CertFreeCertificateContext(cert2);
+        CertCloseStore(store, 0);
 
         res = RegOpenKeyExW(reg_store_saved_certs[i].key, key_name, 0, KEY_ALL_ACCESS, &key);
         ok (res, "The cert's registry entry should be absent at %li, %lx\n", i, GetLastError());
@@ -1394,7 +1317,7 @@
      "Expected E_INVALIDARG, got %08lx\n", GetLastError());
     /* The name is expected to be UNICODE, check with an ASCII name */
     store = CertOpenStore(CERT_STORE_PROV_SYSTEM_REGISTRY, 0, 0,
-     CERT_SYSTEM_STORE_CURRENT_USER | CERT_STORE_OPEN_EXISTING_FLAG, "My\0");
+     CERT_SYSTEM_STORE_CURRENT_USER | CERT_STORE_OPEN_EXISTING_FLAG, "My");
     ok(!store && GetLastError() == ERROR_FILE_NOT_FOUND,
      "Expected ERROR_FILE_NOT_FOUND, got %08lx\n", GetLastError());
 }
@@ -1419,7 +1342,7 @@
      "Expected ERROR_FILE_NOT_FOUND, got %08lx\n", GetLastError());
     /* The name is expected to be UNICODE, first check with an ASCII name */
     store = CertOpenStore(CERT_STORE_PROV_SYSTEM, 0, 0,
-     CERT_SYSTEM_STORE_CURRENT_USER | CERT_STORE_OPEN_EXISTING_FLAG, "My\0");
+     CERT_SYSTEM_STORE_CURRENT_USER | CERT_STORE_OPEN_EXISTING_FLAG, "My");
     ok(!store && GetLastError() == ERROR_FILE_NOT_FOUND,
      "Expected ERROR_FILE_NOT_FOUND, got %08lx\n", GetLastError());
     /* Create the expected key */
@@ -2807,7 +2730,7 @@
 
 static void testEmptyStore(void)
 {
-    const CERT_CONTEXT *cert, *cert2, *cert3, *copy;
+    const CERT_CONTEXT *cert, *cert2, *cert3;
     const CRL_CONTEXT *crl;
     const CTL_CONTEXT *ctl;
     HCERTSTORE store;
@@ -2831,18 +2754,17 @@
 
     res = CertAddCertificateContextToStore(cert->hCertStore, cert2, CERT_STORE_ADD_NEW, &cert3);
     ok(res, "CertAddCertificateContextToStore failed\n");
-    todo_wine_if(cert3)
+    todo_wine
     ok(cert3 && cert3 != cert2, "Unexpected cert3\n");
     ok(cert3->hCertStore == cert->hCertStore, "Unexpected hCertStore\n");
 
     test_store_is_empty(cert->hCertStore);
 
-    copy = CertDuplicateCertificateContext(cert3);
-    ok(copy == cert3, "got %p, %p.\n", copy, cert3);
-    res = CertDeleteCertificateFromStore(copy);
+    res = CertDeleteCertificateFromStore(cert3);
     ok(res, "CertDeleteCertificateContextFromStore failed\n");
-    ok(cert3->hCertStore == cert->hCertStore, "Unexpected hCertStore %p\n", cert3->hCertStore);
-    CertFreeCertificateContext(copy);
+    ok(cert3->hCertStore == cert->hCertStore, "Unexpected hCertStore\n");
+
+    CertFreeCertificateContext(cert3);
 
     store = CertOpenStore(CERT_STORE_PROV_MEMORY, 0, 0, CERT_STORE_CREATE_NEW_FLAG, NULL);
     ok(store != NULL, "CertOpenStore failed\n");
@@ -2852,15 +2774,12 @@
     ok(cert3 && cert3 != cert2, "Unexpected cert3\n");
     ok(cert3->hCertStore == store, "Unexpected hCertStore\n");
 
-    copy = CertDuplicateCertificateContext(cert3);
-    ok(copy == cert3, "got %p, %p.\n", copy, cert3);
-    res = CertDeleteCertificateFromStore(copy);
+    res = CertDeleteCertificateFromStore(cert3);
     ok(res, "CertDeleteCertificateContextFromStore failed\n");
     ok(cert3->hCertStore == store, "Unexpected hCertStore\n");
-    CertFreeCertificateContext(copy);
-
-    res = CertCloseStore(store, CERT_CLOSE_STORE_CHECK_FLAG);
-    ok(res, "got error %#lx.\n", GetLastError());
+
+    CertCloseStore(store, 0);
+    CertFreeCertificateContext(cert3);
 
     res = CertCloseStore(cert->hCertStore, CERT_CLOSE_STORE_CHECK_FLAG);
     ok(!res && GetLastError() == E_UNEXPECTED, "CertCloseStore returned: %x(%lx)\n", res, GetLastError());
@@ -3045,10 +2964,8 @@
     certs = countCertsInStore(store1);
     ok(certs == 0, "Expected 0 certs, got %ld\n", certs);
 
-    ret = CertCloseStore(store1, CERT_CLOSE_STORE_CHECK_FLAG);
-    ok(ret, "got error %#lx.\n", GetLastError());
-    ret = CertCloseStore(store2, CERT_CLOSE_STORE_CHECK_FLAG);
-    ok(ret, "got error %#lx.\n", GetLastError());
+    CertCloseStore(store1, 0);
+    CertCloseStore(store2, 0);
 }
 
 static const BYTE pfxdata[] =
@@ -3370,16 +3287,6 @@
 
     CertFreeCertificateContext( cert );
     CertCloseStore( store, 0 );
-
-    /* PKCS12_NO_PERSIST_KEY|PKCS12_ALWAYS_CNG_KSP */
-    store = PFXImportCertStore( &pfx, NULL, PKCS12_NO_PERSIST_KEY|PKCS12_ALWAYS_CNG_KSP );
-    ok( store != NULL, "got %lu\n", GetLastError() );
-
-    cert = CertFindCertificateInStore( store, X509_ASN_ENCODING, 0, CERT_FIND_ANY, NULL, NULL );
-    ok( cert != NULL, "got %08lx\n", GetLastError() );
-
-    CertFreeCertificateContext( cert );
-    CertCloseStore( store, 0 );
 }
 
 static void test_CryptQueryObject(void)
