diff --git a/dlls/windowscodecs/ddsformat.c b/dlls/windowscodecs/ddsformat.c
--- a/dlls/windowscodecs/ddsformat.c
+++ b/dlls/windowscodecs/ddsformat.c
@@ -141,6 +141,7 @@
 typedef struct DdsDecoder {
     IWICBitmapDecoder IWICBitmapDecoder_iface;
     IWICDdsDecoder IWICDdsDecoder_iface;
+    IWICWineDecoder IWICWineDecoder_iface;
     LONG ref;
     BOOL initialized;
     IStream *stream;
@@ -721,6 +722,11 @@
     return CONTAINING_RECORD(iface, DdsDecoder, IWICDdsDecoder_iface);
 }
 
+static inline DdsDecoder *impl_from_IWICWineDecoder(IWICWineDecoder *iface)
+{
+    return CONTAINING_RECORD(iface, DdsDecoder, IWICWineDecoder_iface);
+}
+
 static inline DdsFrameDecode *impl_from_IWICBitmapFrameDecode(IWICBitmapFrameDecode *iface)
 {
     return CONTAINING_RECORD(iface, DdsFrameDecode, IWICBitmapFrameDecode_iface);
@@ -1065,6 +1071,8 @@
         *ppv = &This->IWICBitmapDecoder_iface;
     } else if (IsEqualIID(&IID_IWICDdsDecoder, iid)) {
         *ppv = &This->IWICDdsDecoder_iface;
+    } else if (IsEqualIID(&IID_IWICWineDecoder, iid)) {
+        *ppv = &This->IWICWineDecoder_iface;
     } else {
         *ppv = NULL;
         return E_NOINTERFACE;
@@ -1110,75 +1118,27 @@
     return E_NOTIMPL;
 }
 
-static HRESULT WINAPI DdsDecoder_Initialize(IWICBitmapDecoder *iface, IStream *stream,
+static HRESULT WINAPI DdsDecoder_Initialize(IWICBitmapDecoder *iface, IStream *pIStream,
                                             WICDecodeOptions cacheOptions)
 {
     DdsDecoder *This = impl_from_IWICBitmapDecoder(iface);
-    DDS_HEADER_DXT10 header_dxt10;
-    LARGE_INTEGER seek;
-    DDS_HEADER header;
-    ULONG bytesread;
-    DWORD magic;
     HRESULT hr;
 
-    TRACE("(%p,%p,%x)\n", iface, stream, cacheOptions);
+    TRACE("(%p,%p,%x)\n", iface, pIStream, cacheOptions);
 
     EnterCriticalSection(&This->lock);
 
-    if (This->initialized) {
-        hr = WINCODEC_ERR_WRONGSTATE;
-        goto end;
-    }
-
-    seek.QuadPart = 0;
-    hr = IStream_Seek(stream, seek, STREAM_SEEK_SET, NULL);
+    hr = IWICWineDecoder_Initialize(&This->IWICWineDecoder_iface, pIStream, cacheOptions);
     if (FAILED(hr)) goto end;
-
-    hr = IStream_Read(stream, &magic, sizeof(magic), &bytesread);
-    if (FAILED(hr)) goto end;
-    if (bytesread != sizeof(magic)) {
-        hr = WINCODEC_ERR_STREAMREAD;
-        goto end;
-    }
-    if (magic != DDS_MAGIC) {
-        hr = WINCODEC_ERR_UNKNOWNIMAGEFORMAT;
-        goto end;
-    }
-
-    hr = IStream_Read(stream, &header, sizeof(header), &bytesread);
-    if (FAILED(hr)) goto end;
-    if (bytesread != sizeof(header)) {
-        hr = WINCODEC_ERR_STREAMREAD;
-        goto end;
-    }
-    if (header.size != sizeof(header)) {
-        hr = WINCODEC_ERR_BADHEADER;
-        goto end;
-    }
-
-    if (has_extended_header(&header)) {
-        hr = IStream_Read(stream, &header_dxt10, sizeof(header_dxt10), &bytesread);
-        if (FAILED(hr)) goto end;
-        if (bytesread != sizeof(header_dxt10)) {
-            hr = WINCODEC_ERR_STREAMREAD;
-            goto end;
-        }
-    }
-
-    get_dds_info(&This->info, &header, &header_dxt10);
 
     if (This->info.dimension == WICDdsTextureCube ||
         (This->info.format != DXGI_FORMAT_BC1_UNORM &&
          This->info.format != DXGI_FORMAT_BC2_UNORM &&
-         This->info.format != DXGI_FORMAT_BC3_UNORM))
-    {
+         This->info.format != DXGI_FORMAT_BC3_UNORM)) {
+        IStream_Release(pIStream);
+        This->stream = NULL;
+        This->initialized = FALSE;
         hr = WINCODEC_ERR_BADHEADER;
-    }
-    else
-    {
-        This->initialized = TRUE;
-        This->stream = stream;
-        IStream_AddRef(stream);
     }
 
 end:
@@ -1445,7 +1405,7 @@
     frame_decode->info.pixel_format_bpp = This->info.pixel_format_bpp;
     frame_decode->block_data = malloc(frame_size);
     frame_decode->pixel_data = NULL;
-    hr = IStream_Seek(This->stream, seek, STREAM_SEEK_SET, NULL);
+    hr = IStream_Seek(This->stream, seek, SEEK_SET, NULL);
     if (hr != S_OK) goto end;
     hr = IStream_Read(This->stream, frame_decode->block_data, frame_size, &bytesread);
     if (hr != S_OK || bytesread != frame_size) {
@@ -1470,6 +1430,97 @@
     DdsDecoder_Dds_Release,
     DdsDecoder_Dds_GetParameters,
     DdsDecoder_Dds_GetFrame
+};
+
+static HRESULT WINAPI DdsDecoder_Wine_QueryInterface(IWICWineDecoder *iface, REFIID iid, void **ppv)
+{
+    DdsDecoder *This = impl_from_IWICWineDecoder(iface);
+    return DdsDecoder_QueryInterface(&This->IWICBitmapDecoder_iface, iid, ppv);
+}
+
+static ULONG WINAPI DdsDecoder_Wine_AddRef(IWICWineDecoder *iface)
+{
+    DdsDecoder *This = impl_from_IWICWineDecoder(iface);
+    return DdsDecoder_AddRef(&This->IWICBitmapDecoder_iface);
+}
+
+static ULONG WINAPI DdsDecoder_Wine_Release(IWICWineDecoder *iface)
+{
+    DdsDecoder *This = impl_from_IWICWineDecoder(iface);
+    return DdsDecoder_Release(&This->IWICBitmapDecoder_iface);
+}
+
+static HRESULT WINAPI DdsDecoder_Wine_Initialize(IWICWineDecoder *iface, IStream *stream, WICDecodeOptions options)
+{
+    DdsDecoder *This = impl_from_IWICWineDecoder(iface);
+    DDS_HEADER_DXT10 header_dxt10;
+    LARGE_INTEGER seek;
+    DDS_HEADER header;
+    ULONG bytesread;
+    DWORD magic;
+    HRESULT hr;
+
+    TRACE("(This %p, stream %p, options %#x)\n", iface, stream, options);
+
+    EnterCriticalSection(&This->lock);
+
+    if (This->initialized) {
+        hr = WINCODEC_ERR_WRONGSTATE;
+        goto end;
+    }
+
+    seek.QuadPart = 0;
+    hr = IStream_Seek(stream, seek, SEEK_SET, NULL);
+    if (FAILED(hr)) goto end;
+
+    hr = IStream_Read(stream, &magic, sizeof(magic), &bytesread);
+    if (FAILED(hr)) goto end;
+    if (bytesread != sizeof(magic)) {
+        hr = WINCODEC_ERR_STREAMREAD;
+        goto end;
+    }
+    if (magic != DDS_MAGIC) {
+        hr = WINCODEC_ERR_UNKNOWNIMAGEFORMAT;
+        goto end;
+    }
+
+    hr = IStream_Read(stream, &header, sizeof(header), &bytesread);
+    if (FAILED(hr)) goto end;
+    if (bytesread != sizeof(header)) {
+        hr = WINCODEC_ERR_STREAMREAD;
+        goto end;
+    }
+    if (header.size != sizeof(header)) {
+        hr = WINCODEC_ERR_BADHEADER;
+        goto end;
+    }
+
+    if (has_extended_header(&header)) {
+        hr = IStream_Read(stream, &header_dxt10, sizeof(header_dxt10), &bytesread);
+        if (FAILED(hr)) goto end;
+        if (bytesread != sizeof(header_dxt10)) {
+            hr = WINCODEC_ERR_STREAMREAD;
+            goto end;
+        }
+    }
+
+    get_dds_info(&This->info, &header, &header_dxt10);
+
+    This->initialized = TRUE;
+    This->stream = stream;
+    IStream_AddRef(stream);
+
+end:
+    LeaveCriticalSection(&This->lock);
+
+    return hr;
+}
+
+static const IWICWineDecoderVtbl DdsDecoder_Wine_Vtbl = {
+    DdsDecoder_Wine_QueryInterface,
+    DdsDecoder_Wine_AddRef,
+    DdsDecoder_Wine_Release,
+    DdsDecoder_Wine_Initialize
 };
 
 static HRESULT WINAPI DdsFrameEncode_QueryInterface(IWICBitmapFrameEncode *iface, REFIID iid, void **ppv)
@@ -1706,6 +1757,7 @@
 
     This->IWICBitmapDecoder_iface.lpVtbl = &DdsDecoder_Vtbl;
     This->IWICDdsDecoder_iface.lpVtbl = &DdsDecoder_Dds_Vtbl;
+    This->IWICWineDecoder_iface.lpVtbl = &DdsDecoder_Wine_Vtbl;
     This->ref = 1;
     This->initialized = FALSE;
     This->stream = NULL;
