diff --git a/dlls/scrrun/tests/filesystem.c b/dlls/scrrun/tests/filesystem.c
--- a/dlls/scrrun/tests/filesystem.c
+++ b/dlls/scrrun/tests/filesystem.c
@@ -1738,76 +1738,6 @@
     SysFreeString(nameW);
 }
 
-static HRESULT write_blank_lines(WCHAR *nameW, VARIANT_BOOL is_unicode, LONG lines)
-{
-    HRESULT hr;
-    ITextStream *stream;
-
-    hr = IFileSystem3_CreateTextFile(fs3, nameW, VARIANT_FALSE, is_unicode, &stream);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = ITextStream_WriteBlankLines(stream, lines);
-    ITextStream_Release(stream);
-    return hr;
-}
-
-static void test_WriteBlankLines(void)
-{
-    WCHAR pathW[MAX_PATH], dirW[MAX_PATH];
-    BSTR nameW;
-    HRESULT hr;
-    BOOL ret;
-
-    get_temp_filepath(testfileW, pathW, dirW);
-
-    ret = CreateDirectoryW(dirW, NULL);
-    ok(ret, "Unexpected retval %d, error %ld.\n", ret, GetLastError());
-
-    nameW = SysAllocString(pathW);
-
-    hr = write_blank_lines(nameW, VARIANT_FALSE, -1);
-    ok(hr == E_INVALIDARG, "Unexpected hr %#lx.\n", hr);
-    test_file_contents(pathW, 0, NULL);
-    DeleteFileW(nameW);
-
-    hr = write_blank_lines(nameW, VARIANT_FALSE, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    test_file_contents(pathW, 0, NULL);
-    DeleteFileW(nameW);
-
-    hr = write_blank_lines(nameW, VARIANT_FALSE, 1);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    test_file_contents(pathW, 2, "\r\n");
-    DeleteFileW(nameW);
-
-    hr = write_blank_lines(nameW, VARIANT_FALSE, 2);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    test_file_contents(pathW, 4, "\r\n\r\n");
-    DeleteFileW(nameW);
-
-    hr = write_blank_lines(nameW, VARIANT_TRUE, -1);
-    ok(hr == E_INVALIDARG, "Unexpected hr %#lx.\n", hr);
-    test_file_contents(pathW, 1*sizeof(WCHAR), L"\ufeff");
-    DeleteFileW(nameW);
-
-    hr = write_blank_lines(nameW, VARIANT_TRUE, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    test_file_contents(pathW, 1*sizeof(WCHAR), L"\ufeff");
-    DeleteFileW(nameW);
-
-    hr = write_blank_lines(nameW, VARIANT_TRUE, 1);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    test_file_contents(pathW, 3*sizeof(WCHAR), L"\ufeff\r\n");
-    DeleteFileW(nameW);
-
-    hr = write_blank_lines(nameW, VARIANT_TRUE, 2);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    test_file_contents(pathW, 5*sizeof(WCHAR), L"\ufeff\r\n\r\n");
-    DeleteFileW(nameW);
-
-    RemoveDirectoryW(dirW);
-    SysFreeString(nameW);
-}
-
 static void test_ReadAll(void)
 {
     static const WCHAR firstlineW[] = L"first";
@@ -2707,135 +2637,6 @@
     hr = IFileSystem3_MoveFile(fs3, NULL, str);
     ok(hr == E_INVALIDARG, "Unexpected hr %#lx.\n", hr);
     SysFreeString(str);
-}
-
-static void test_MoveFolder(void)
-{
-    WCHAR temp_path[MAX_PATH], cwd[MAX_PATH];
-    BSTR src, dst, empty;
-    HANDLE file;
-    HRESULT hr;
-    BOOL ret;
-
-    GetCurrentDirectoryW(MAX_PATH, cwd);
-    GetTempPathW(MAX_PATH, temp_path);
-    SetCurrentDirectoryW(temp_path);
-    src = SysAllocString(L"winetest_src");
-    dst = SysAllocString(L"winetest_dst");
-
-    ret = CreateDirectoryW(L"winetest_src", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_PATHNOTFOUND, "Got hr %#lx.\n", hr);
-    ret = CreateDirectoryW(L"winetest_src", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_FILEALREADYEXISTS, "Got hr %#lx.\n", hr);
-    ret = RemoveDirectoryW(L"winetest_dst");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = RemoveDirectoryW(L"winetest_src");
-    ok(ret, "Got error %lu.\n", GetLastError());
-
-    empty = SysAllocString(L"");
-    hr = IFileSystem3_MoveFolder(fs3, src, NULL);
-    ok(hr == E_INVALIDARG, "Got hr %#lx.\n", hr);
-    hr = IFileSystem3_MoveFolder(fs3, NULL, dst);
-    ok(hr == E_INVALIDARG, "Got hr %#lx.\n", hr);
-    hr = IFileSystem3_MoveFolder(fs3, src, empty);
-    ok(hr == E_INVALIDARG, "Got hr %#lx.\n", hr);
-    hr = IFileSystem3_MoveFolder(fs3, empty, dst);
-    ok(hr == E_INVALIDARG, "Got hr %#lx.\n", hr);
-    SysFreeString(empty);
-
-    file = CreateFileW(L"winetest_src", GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
-    ok(file != INVALID_HANDLE_VALUE, "Got error %lu.\n", GetLastError());
-    CloseHandle(file);
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_PATHNOTFOUND, "Got hr %#lx.\n", hr);
-    ret = DeleteFileW(L"winetest_src");
-    ok(ret, "Got error %lu.\n", GetLastError());
-
-    dst = SysAllocString(L"winetest_dst\\");
-
-    ret = CreateDirectoryW(L"winetest_src", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_PATHNOTFOUND, "Got hr %#lx.\n", hr);
-    ret = CreateDirectoryW(L"winetest_dst", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    ret = RemoveDirectoryW(L"winetest_dst\\winetest_src");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = RemoveDirectoryW(L"winetest_dst");
-    ok(ret, "Got error %lu.\n", GetLastError());
-
-    src = SysAllocString(L"winetest_src\\");
-    dst = SysAllocString(L"winetest_dst");
-
-    ret = CreateDirectoryW(L"winetest_src", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_ILLEGALFUNCTIONCALL, "Got hr %#lx.\n", hr);
-    src = SysAllocString(L"winetest_src\\");
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_ILLEGALFUNCTIONCALL, "Got hr %#lx.\n", hr);
-
-    ret = RemoveDirectoryW(L"winetest_src");
-    ok(ret, "Got error %lu.\n", GetLastError());
-
-    ret = CreateDirectoryW(L"winetest_src1", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = CreateDirectoryW(L"winetest_src2", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    file = CreateFileW(L"winetest_src3", GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
-    ok(file != INVALID_HANDLE_VALUE, "Got error %lu.\n", GetLastError());
-    CloseHandle(file);
-
-    src = SysAllocString(L"winetest_src*");
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_PATHNOTFOUND, "Got hr %#lx.\n", hr);
-    ret = CreateDirectoryW(L"winetest_dst", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_PATHNOTFOUND, "Got hr %#lx.\n", hr);
-    ret = RemoveDirectoryW(L"winetest_dst/winetest_src1");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = RemoveDirectoryW(L"winetest_dst/winetest_src2");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = RemoveDirectoryW(L"winetest_dst");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = DeleteFileW(L"winetest_src3");
-    ok(ret, "Got error %lu.\n", GetLastError());
-
-    /* Fail to copy one file. */
-    ret = CreateDirectoryW(L"winetest_src1", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = CreateDirectoryW(L"winetest_src2", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = CreateDirectoryW(L"winetest_src3", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = CreateDirectoryW(L"winetest_dst", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = CreateDirectoryW(L"winetest_dst\\winetest_src2", NULL);
-    ok(ret, "Got error %lu.\n", GetLastError());
-    hr = IFileSystem3_MoveFolder(fs3, src, dst);
-    ok(hr == CTL_E_FILEALREADYEXISTS, "Got hr %#lx.\n", hr);
-    ret = RemoveDirectoryW(L"winetest_dst/winetest_src1");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = RemoveDirectoryW(L"winetest_dst/winetest_src2");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = RemoveDirectoryW(L"winetest_dst");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = RemoveDirectoryW(L"winetest_src2");
-    ok(ret, "Got error %lu.\n", GetLastError());
-    ret = RemoveDirectoryW(L"winetest_src3");
-    ok(ret, "Got error %lu.\n", GetLastError());
 }
 
 static void test_DoOpenPipeStream(void)
@@ -2961,7 +2762,6 @@
     test_CreateTextFile();
     test_FolderCreateTextFile();
     test_WriteLine();
-    test_WriteBlankLines();
     test_ReadAll();
     test_Read();
     test_ReadLine();
@@ -2972,7 +2772,6 @@
     test_GetExtensionName();
     test_GetSpecialFolder();
     test_MoveFile();
-    test_MoveFolder();
     test_DoOpenPipeStream();
 
     IFileSystem3_Release(fs3);
