diff --git a/dlls/comdlg32/itemdlg.c b/dlls/comdlg32/itemdlg.c
--- a/dlls/comdlg32/itemdlg.c
+++ b/dlls/comdlg32/itemdlg.c
@@ -498,7 +498,7 @@
     if(FAILED(hr) || !item_count)
         return;
 
-    names = malloc(item_count*sizeof(LPWSTR));
+    names = HeapAlloc(GetProcessHeap(), 0, item_count*sizeof(LPWSTR));
 
     /* Get names of the selected items */
     valid_count = 0; len_total = 0;
@@ -532,7 +532,7 @@
     }
     else if(valid_count > 1)
     {
-        LPWSTR string = malloc(sizeof(WCHAR)*len_total);
+        LPWSTR string = HeapAlloc(GetProcessHeap(), 0, sizeof(WCHAR)*len_total);
         LPWSTR cur_point = string;
 
         for(i = 0; i < valid_count; i++)
@@ -548,10 +548,10 @@
         *(cur_point-1) = '\0';
 
         set_file_name(This, string);
-        free(string);
-    }
-
-    free(names);
+        HeapFree(GetProcessHeap(), 0, string);
+    }
+
+    HeapFree(GetProcessHeap(), 0, names);
     return;
 }
 
@@ -622,13 +622,13 @@
     if(FAILED(hr))
     {
         ERR("Failed to get pidl for current directory.\n");
-        free(files);
+        HeapFree(GetProcessHeap(), 0, files);
         return hr;
     }
 
     TRACE("Acting on %d file(s).\n", file_count);
 
-    pidla = malloc(sizeof(LPITEMIDLIST) * file_count);
+    pidla = HeapAlloc(GetProcessHeap(), 0, sizeof(LPITEMIDLIST) * file_count);
     open_action = ONOPEN_OPEN;
     fn_iter = files;
 
@@ -794,11 +794,11 @@
     }
 
     /* Clean up */
-    free(files);
+    HeapFree(GetProcessHeap(), 0, files);
     ILFree(current_folder);
     for(i = 0; i < file_count; i++)
         ILFree(pidla[i]);
-    free(pidla);
+    HeapFree(GetProcessHeap(), 0, pidla);
 
     /* Success closes the dialog */
     return ret;
@@ -840,8 +840,8 @@
 static void item_free(cctrl_item *item)
 {
     DestroyWindow(item->hwnd);
-    free(item->label);
-    free(item);
+    HeapFree(GetProcessHeap(), 0, item->label);
+    HeapFree(GetProcessHeap(), 0, item);
 }
 
 static cctrl_item* get_item(customctrl* parent, DWORD itemid, CDCONTROLSTATEF visible_flags, DWORD* position)
@@ -886,13 +886,13 @@
     if (get_item(parent, itemid, 0, NULL))
         return E_INVALIDARG;
 
-    item = malloc(sizeof(*item));
-    label_copy = malloc((lstrlenW(label)+1)*sizeof(WCHAR));
+    item = HeapAlloc(GetProcessHeap(), 0, sizeof(*item));
+    label_copy = HeapAlloc(GetProcessHeap(), 0, (lstrlenW(label)+1)*sizeof(WCHAR));
 
     if (!item || !label_copy)
     {
-        free(item);
-        free(label_copy);
+        HeapFree(GetProcessHeap(), 0, item);
+        HeapFree(GetProcessHeap(), 0, label_copy);
         return E_OUTOFMEMORY;
     }
 
@@ -965,7 +965,7 @@
     TRACE("\n");
 
     len = SendMessageW(hctrl, WM_GETTEXTLENGTH, 0, 0);
-    text = malloc(sizeof(WCHAR)*(len+1));
+    text = HeapAlloc(GetProcessHeap(), 0, sizeof(WCHAR)*(len+1));
     if(!text) return;
     SendMessageW(hctrl, WM_GETTEXT, len+1, (LPARAM)text);
 
@@ -994,7 +994,7 @@
     SetWindowPos(hctrl, NULL, 0, 0, final_width, final_height,
                  SWP_NOZORDER | SWP_NOMOVE | SWP_NOACTIVATE);
 
-    free(text);
+    HeapFree(GetProcessHeap(), 0, text);
 }
 
 static UINT ctrl_get_height(customctrl *ctrl) {
@@ -1029,7 +1029,7 @@
     }
 
     DestroyWindow(ctrl->hwnd);
-    free(ctrl);
+    HeapFree(GetProcessHeap(), 0, ctrl);
 }
 
 static void customctrl_resize(FileDialogImpl *This, customctrl *ctrl)
@@ -1267,7 +1267,7 @@
 
     SetPropW(ns_hwnd, L"nfs_child", control_hwnd);
 
-    ctrl = calloc(1, sizeof(customctrl));
+    ctrl = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(customctrl));
     if(!ctrl)
         return E_OUTOFMEMORY;
 
@@ -1454,7 +1454,7 @@
         SetWindowLongW(This->cctrls_hwnd, GWL_STYLE, wndstyle);
 
         SetParent(This->cctrls_hwnd, parent);
-        ShowWindow(This->cctrls_hwnd, SW_SHOW);
+        ShowWindow(This->cctrls_hwnd, TRUE);
 
         /* Set the fonts to match the dialog font. */
         font = (HFONT)SendMessageW(parent, WM_GETFONT, 0, 0);
@@ -1469,7 +1469,7 @@
     }
     else
     {
-        ShowWindow(This->cctrls_hwnd, SW_HIDE);
+        ShowWindow(This->cctrls_hwnd, FALSE);
 
         wndstyle = GetWindowLongW(This->cctrls_hwnd, GWL_STYLE);
         wndstyle &= ~(WS_CHILD);
@@ -2420,7 +2420,6 @@
 
     *ppvObject = NULL;
     if(IsEqualGUID(riid, &IID_IUnknown) ||
-       IsEqualGUID(riid, &IID_IModalWindow) ||
        IsEqualGUID(riid, &IID_IFileDialog) ||
        IsEqualGUID(riid, &IID_IFileDialog2))
     {
@@ -2492,7 +2491,7 @@
             LocalFree((void*)This->filterspecs[i].pszName);
             LocalFree((void*)This->filterspecs[i].pszSpec);
         }
-        free(This->filterspecs);
+        HeapFree(GetProcessHeap(), 0, This->filterspecs);
 
         DestroyWindow(This->cctrls_hwnd);
 
@@ -2513,7 +2512,7 @@
         DestroyMenu(This->hmenu_opendropdown);
         DeleteObject(This->hfont_opendropdown);
 
-        free(This);
+        HeapFree(GetProcessHeap(), 0, This);
     }
 
     return ref;
@@ -2545,31 +2544,11 @@
     if(!cFileTypes)
         return S_OK;
 
-    This->filterspecs = malloc(sizeof(COMDLG_FILTERSPEC)*cFileTypes);
+    This->filterspecs = HeapAlloc(GetProcessHeap(), 0, sizeof(COMDLG_FILTERSPEC)*cFileTypes);
     for(i = 0; i < cFileTypes; i++)
     {
         This->filterspecs[i].pszName = StrDupW(rgFilterSpec[i].pszName);
         This->filterspecs[i].pszSpec = StrDupW(rgFilterSpec[i].pszSpec);
-
-        if (This->filterspecs[i].pszName != NULL && This->filterspecs[i].pszSpec != NULL)
-        {
-            DWORD name_len = lstrlenW(This->filterspecs[i].pszName);
-
-            if (name_len == 0 || This->filterspecs[i].pszName[name_len - 1] != L')')
-            {
-                DWORD spec_len = lstrlenW(This->filterspecs[i].pszSpec);
-
-                DWORD total_len = name_len + spec_len + 4;
-
-                WCHAR* pszName = LocalAlloc(LMEM_FIXED, total_len * sizeof(WCHAR));
-                if (pszName != NULL)
-                {
-                    swprintf(pszName, total_len, L"%s (%s)", This->filterspecs[i].pszName, This->filterspecs[i].pszSpec);
-                    LocalFree((void *)This->filterspecs[i].pszName);
-                    This->filterspecs[i].pszName = pszName;
-                }
-            }
-        }
     }
     This->filterspec_count = cFileTypes;
 
@@ -2620,7 +2599,7 @@
     if(!pfde || !pdwCookie)
         return E_INVALIDARG;
 
-    client = malloc(sizeof(events_client));
+    client = HeapAlloc(GetProcessHeap(), 0, sizeof(events_client));
     client->pfde = pfde;
     client->cookie = ++This->events_next_cookie;
 
@@ -2651,7 +2630,7 @@
     {
         list_remove(&found->entry);
         IFileDialogEvents_Release(found->pfde);
-        free(found);
+        HeapFree(GetProcessHeap(), 0, found);
         return S_OK;
     }
 
@@ -4768,7 +4747,7 @@
     if(pUnkOuter)
         return CLASS_E_NOAGGREGATION;
 
-    fdimpl = calloc(1, sizeof(FileDialogImpl));
+    fdimpl = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(FileDialogImpl));
     if(!fdimpl)
         return E_OUTOFMEMORY;
 
