diff --git a/dlls/d3dx10_43/async.c b/dlls/d3dx10_43/async.c
--- a/dlls/d3dx10_43/async.c
+++ b/dlls/d3dx10_43/async.c
@@ -88,36 +88,6 @@
     memorydataloader_Destroy
 };
 
-HRESULT load_file(const WCHAR *path, void **data, DWORD *size)
-{
-    DWORD read_len;
-    HANDLE file;
-    BOOL ret;
-
-    file = CreateFileW(path, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
-            NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
-    if (file == INVALID_HANDLE_VALUE)
-        return D3D10_ERROR_FILE_NOT_FOUND;
-
-    *size = GetFileSize(file, NULL);
-    *data = malloc(*size);
-    if (!*data)
-    {
-        CloseHandle(file);
-        return E_OUTOFMEMORY;
-    }
-
-    ret = ReadFile(file, *data, *size, &read_len, NULL);
-    CloseHandle(file);
-    if (!ret || read_len != *size)
-    {
-        WARN("Failed to read file contents.\n");
-        free(*data);
-        return E_FAIL;
-    }
-    return S_OK;
-}
-
 static HRESULT WINAPI filedataloader_Load(ID3DX10DataLoader *iface)
 {
     struct asyncdataloader *loader = impl_from_ID3DX10DataLoader(iface);
@@ -173,63 +143,6 @@
     filedataloader_Destroy
 };
 
-static HRESULT load_resource_initA(HMODULE module, const char *resource, HRSRC *rsrc)
-{
-    if (!(*rsrc = FindResourceA(module, resource, (const char *)RT_RCDATA)))
-        *rsrc = FindResourceA(module, resource, (const char *)RT_BITMAP);
-    if (!*rsrc)
-    {
-        WARN("Failed to find resource.\n");
-        return D3DX10_ERR_INVALID_DATA;
-    }
-    return S_OK;
-}
-
-static HRESULT load_resource_initW(HMODULE module, const WCHAR *resource, HRSRC *rsrc)
-{
-    if (!(*rsrc = FindResourceW(module, resource, (const WCHAR *)RT_RCDATA)))
-        *rsrc = FindResourceW(module, resource, (const WCHAR *)RT_BITMAP);
-    if (!*rsrc)
-    {
-        WARN("Failed to find resource.\n");
-        return D3DX10_ERR_INVALID_DATA;
-    }
-    return S_OK;
-}
-
-static HRESULT load_resource(HMODULE module, HRSRC rsrc, void **data, DWORD *size)
-{
-    HGLOBAL hglobal;
-
-    if (!(*size = SizeofResource(module, rsrc)))
-        return D3DX10_ERR_INVALID_DATA;
-    if (!(hglobal = LoadResource(module, rsrc)))
-        return D3DX10_ERR_INVALID_DATA;
-    if (!(*data = LockResource(hglobal)))
-        return D3DX10_ERR_INVALID_DATA;
-    return S_OK;
-}
-
-HRESULT load_resourceA(HMODULE module, const char *resource, void **data, DWORD *size)
-{
-    HRESULT hr;
-    HRSRC rsrc;
-
-    if (FAILED((hr = load_resource_initA(module, resource, &rsrc))))
-        return hr;
-    return load_resource(module, rsrc, data, size);
-}
-
-HRESULT load_resourceW(HMODULE module, const WCHAR *resource, void **data, DWORD *size)
-{
-    HRESULT hr;
-    HRSRC rsrc;
-
-    if ((FAILED(hr = load_resource_initW(module, resource, &rsrc))))
-        return hr;
-    return load_resource(module, rsrc, data, size);
-}
-
 static HRESULT WINAPI resourcedataloader_Load(ID3DX10DataLoader *iface)
 {
     struct asyncdataloader *loader = impl_from_ID3DX10DataLoader(iface);
@@ -323,7 +236,6 @@
     ID3DX10DataProcessor ID3DX10DataProcessor_iface;
     ID3D10Device *device;
     D3DX10_IMAGE_INFO img_info;
-    D3DX10_IMAGE_INFO *img_info_out;
     D3DX10_IMAGE_LOAD_INFO load_info;
     D3D10_SUBRESOURCE_DATA *resource_data;
 };
@@ -336,7 +248,6 @@
 static HRESULT WINAPI texture_processor_Process(ID3DX10DataProcessor *iface, void *data, SIZE_T size)
 {
     struct texture_processor *processor = texture_processor_from_ID3DX10DataProcessor(iface);
-    HRESULT hr;
 
     TRACE("iface %p, data %p, size %Iu.\n", iface, data, size);
 
@@ -346,10 +257,7 @@
         free(processor->resource_data);
         processor->resource_data = NULL;
     }
-    hr = load_texture_data(data, size, &processor->load_info, &processor->resource_data);
-    if (SUCCEEDED(hr) && processor->img_info_out)
-        *processor->img_info_out = processor->img_info;
-    return hr;
+    return load_texture_data(data, size, &processor->load_info, &processor->resource_data);
 }
 
 static HRESULT WINAPI texture_processor_CreateDeviceObject(ID3DX10DataProcessor *iface, void **object)
@@ -382,6 +290,63 @@
     texture_processor_Process,
     texture_processor_CreateDeviceObject,
     texture_processor_Destroy
+};
+
+struct srv_processor
+{
+    ID3DX10DataProcessor ID3DX10DataProcessor_iface;
+    ID3DX10DataProcessor *texture_processor;
+};
+
+static inline struct srv_processor *srv_processor_from_ID3DX10DataProcessor(ID3DX10DataProcessor *iface)
+{
+    return CONTAINING_RECORD(iface, struct srv_processor, ID3DX10DataProcessor_iface);
+}
+
+static HRESULT WINAPI srv_processor_Process(ID3DX10DataProcessor *iface, void *data, SIZE_T size)
+{
+    struct srv_processor *processor = srv_processor_from_ID3DX10DataProcessor(iface);
+
+    TRACE("iface %p, data %p, size %Iu.\n", iface, data, size);
+
+    return ID3DX10DataProcessor_Process(processor->texture_processor, data, size);
+}
+
+static HRESULT WINAPI srv_processor_CreateDeviceObject(ID3DX10DataProcessor *iface, void **object)
+{
+    struct srv_processor *processor = srv_processor_from_ID3DX10DataProcessor(iface);
+    struct texture_processor *tex_processor = texture_processor_from_ID3DX10DataProcessor(processor->texture_processor);
+    ID3D10Resource *texture_resource;
+    HRESULT hr;
+
+    TRACE("iface %p, object %p.\n", iface, object);
+
+    hr = ID3DX10DataProcessor_CreateDeviceObject(processor->texture_processor, (void **)&texture_resource);
+    if (FAILED(hr))
+        return hr;
+
+    hr = ID3D10Device_CreateShaderResourceView(tex_processor->device, texture_resource, NULL,
+            (ID3D10ShaderResourceView **)object);
+    ID3D10Resource_Release(texture_resource);
+    return hr;
+}
+
+static HRESULT WINAPI srv_processor_Destroy(ID3DX10DataProcessor *iface)
+{
+    struct srv_processor *processor = srv_processor_from_ID3DX10DataProcessor(iface);
+
+    TRACE("iface %p.\n", iface);
+
+    ID3DX10DataProcessor_Destroy(processor->texture_processor);
+    free(processor);
+    return S_OK;
+}
+
+static ID3DX10DataProcessorVtbl srv_processor_vtbl =
+{
+    srv_processor_Process,
+    srv_processor_CreateDeviceObject,
+    srv_processor_Destroy
 };
 
 HRESULT WINAPI D3DX10CompileFromMemory(const char *data, SIZE_T data_size, const char *filename,
@@ -600,11 +565,36 @@
     object->ID3DX10DataProcessor_iface.lpVtbl = &texture_processor_vtbl;
     object->device = device;
     ID3D10Device_AddRef(device);
-    if (load_info)
-        object->img_info_out = load_info->pSrcInfo;
     init_load_info(load_info, &object->load_info);
-    object->load_info.pSrcInfo = &object->img_info;
-
+    if (!object->load_info.pSrcInfo)
+        object->load_info.pSrcInfo = &object->img_info;
+
+    *processor = &object->ID3DX10DataProcessor_iface;
+    return S_OK;
+}
+
+HRESULT WINAPI D3DX10CreateAsyncShaderResourceViewProcessor(ID3D10Device *device,
+        D3DX10_IMAGE_LOAD_INFO *load_info, ID3DX10DataProcessor **processor)
+{
+    struct srv_processor *object;
+    HRESULT hr;
+
+    TRACE("device %p, load_info %p, processor %p.\n", device, load_info, processor);
+
+    if (!device || !processor)
+        return E_INVALIDARG;
+
+    object = calloc(1, sizeof(*object));
+    if (!object)
+        return E_OUTOFMEMORY;
+
+    hr = D3DX10CreateAsyncTextureProcessor(device, load_info, &object->texture_processor);
+    if (FAILED(hr))
+    {
+        free(object);
+        return hr;
+    }
+    object->ID3DX10DataProcessor_iface.lpVtbl = &srv_processor_vtbl;
     *processor = &object->ID3DX10DataProcessor_iface;
     return S_OK;
 }
