diff --git a/dlls/msvcp90/locale.c b/dlls/msvcp90/locale.c
--- a/dlls/msvcp90/locale.c
+++ b/dlls/msvcp90/locale.c
@@ -32,6 +32,7 @@
 #include "winbase.h"
 #include "winnls.h"
 #include "msvcp90.h"
+#include "wine/heap.h"
 #include "wine/list.h"
 #include "wine/debug.h"
 
@@ -49,11 +50,14 @@
 wchar_t* __cdecl _W_Getdays(void);
 char* __cdecl _Getmonths(void);
 wchar_t* __cdecl _W_Getmonths(void);
+void* __cdecl _Gettnames(void);
 int __cdecl ___lc_collate_cp_func(void);
 const locale_facet* __thiscall locale__Getfacet(const locale*, size_t);
 const locale* __cdecl locale_classic(void);
 
-#if _MSVCP_VER < 110
+#if _MSVCP_VER >= 110
+wchar_t ** __cdecl ___lc_locale_name_func(void);
+#else
 LCID* __cdecl ___lc_handle_func(void);
 #endif
 
@@ -4007,9 +4011,9 @@
 /* ??0?$codecvt@_SDU_Mbstatet@@@std@@QEAA@AEBV_Locinfo@1@KW4_Codecvt_mode@1@_K@Z */
 DEFINE_THISCALL_WRAPPER(codecvt_char16_ctor_mode, 16)
 codecvt_char16* __thiscall codecvt_char16_ctor_mode(codecvt_char16 *this, const _Locinfo *locinfo,
-        ULONG max_code, codecvt_convert_mode mode, size_t refs)
-{
-    TRACE("(%p %ld %d %Iu)\n", this, max_code, mode, refs);
+        ULONG max_code, codecvt_convert_mode mode, unsigned int refs)
+{
+    TRACE("(%p %ld %d %u)\n", this, max_code, mode, refs);
 
     codecvt_base_ctor_refs(&this->base, refs);
     this->base.facet.vtable = &codecvt_char16_vtable;
@@ -4023,9 +4027,9 @@
 /* ??0?$codecvt@_SDU_Mbstatet@@@std@@QEAA@AEBV_Locinfo@1@_K@Z */
 DEFINE_THISCALL_WRAPPER(codecvt_char16_ctor_locinfo, 12)
 codecvt_char16* __thiscall codecvt_char16_ctor_locinfo(codecvt_char16 *this,
-        const _Locinfo *locinfo, size_t refs)
-{
-    TRACE("(%p %p %Iu)\n", this, locinfo, refs);
+        const _Locinfo *locinfo, unsigned int refs)
+{
+    TRACE("(%p %p %u)\n", this, locinfo, refs);
     return codecvt_char16_ctor_mode(this, locinfo, MAX_UCSCHAR, consume_header, refs);
 }
 
@@ -4033,9 +4037,9 @@
 /* ??0?$codecvt@_SDU_Mbstatet@@@std@@QAE@I@Z */
 /* ??0?$codecvt@_SDU_Mbstatet@@@std@@QEAA@_K@Z */
 DEFINE_THISCALL_WRAPPER(codecvt_char16_ctor_refs, 8)
-codecvt_char16* __thiscall codecvt_char16_ctor_refs(codecvt_char16 *this, size_t refs)
-{
-    TRACE("(%p %Iu)\n", this, refs);
+codecvt_char16* __thiscall codecvt_char16_ctor_refs(codecvt_char16 *this, unsigned int refs)
+{
+    TRACE("(%p %u)\n", this, refs);
     return codecvt_char16_ctor_locinfo(this, NULL, refs);
 }
 
@@ -13071,14 +13075,14 @@
 
     len = MultiByteToWideChar(cv.page, MB_ERR_INVALID_CHARS, src, src_len, NULL, 0);
     if (!len) return INT_MAX;
-    buf = malloc(len * sizeof(WCHAR));
+    buf = heap_alloc(len * sizeof(WCHAR));
     if (!buf) return INT_MAX;
     MultiByteToWideChar(cv.page, MB_ERR_INVALID_CHARS, src, src_len, buf, len);
 
     len = LCMapStringW(lcid, LCMAP_SORTKEY, buf, len, NULL, 0);
     if (len <= dest_len)
         LCMapStringW(lcid, LCMAP_SORTKEY, buf, len, (WCHAR*)dest, dest_len);
-    free(buf);
+    heap_free(buf);
     return len;
 }
 
@@ -13119,43 +13123,43 @@
     return len;
 }
 
-DEFINE_RTTI_DATA(_Facet_base, 0, ".?AV_Facet_base@std@@")
-DEFINE_RTTI_DATA(locale_facet, 0, ".?AVfacet@locale@std@@")
-DEFINE_RTTI_DATA(locale__Locimp, 0, ".?AV_Locimp@locale@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(collate_char, 0, ".?AV?$collate@D@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(collate_wchar, 0, ".?AV?$collate@_W@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(collate_short, 0, ".?AV?$collate@G@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(ctype_base, 0, ".?AUctype_base@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(ctype_char, 0, ".?AV?$ctype@D@std@@", ctype_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(ctype_wchar, 0, ".?AV?$ctype@_W@std@@", ctype_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(ctype_short, 0, ".?AV?$ctype@G@std@@", ctype_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(codecvt_base, 0, ".?AVcodecvt_base@std@@", locale_facet_rtti_base_descriptor)
+DEFINE_RTTI_DATA0(_Facet_base, 0, ".?AV_Facet_base@std@@")
+DEFINE_RTTI_DATA0(locale_facet, 0, ".?AVfacet@locale@std@@")
+DEFINE_RTTI_DATA1(locale__Locimp, 0, &locale_facet_rtti_base_descriptor, ".?AV_Locimp@locale@std@@")
+DEFINE_RTTI_DATA1(collate_char, 0, &locale_facet_rtti_base_descriptor, ".?AV?$collate@D@std@@")
+DEFINE_RTTI_DATA1(collate_wchar, 0, &locale_facet_rtti_base_descriptor, ".?AV?$collate@_W@std@@")
+DEFINE_RTTI_DATA1(collate_short, 0, &locale_facet_rtti_base_descriptor, ".?AV?$collate@G@std@@")
+DEFINE_RTTI_DATA1(ctype_base, 0, &locale_facet_rtti_base_descriptor, ".?AUctype_base@std@@")
+DEFINE_RTTI_DATA2(ctype_char, 0, &ctype_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$ctype@D@std@@")
+DEFINE_RTTI_DATA2(ctype_wchar, 0, &ctype_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$ctype@_W@std@@")
+DEFINE_RTTI_DATA2(ctype_short, 0, &ctype_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$ctype@G@std@@")
+DEFINE_RTTI_DATA1(codecvt_base, 0, &locale_facet_rtti_base_descriptor, ".?AVcodecvt_base@std@@")
 #if _MSVCP_VER >= 140
-DEFINE_RTTI_DATA(codecvt_char, 0, ".?AV?$codecvt@DDU_Mbstatet@@@std@@", codecvt_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(codecvt_char16, 0, ".?AV?$codecvt@_SDU@std@@", codecvt_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(codecvt_char32, 0, ".?AV?$codecvt@_UDU@std@@", codecvt_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(codecvt_wchar, 0, ".?AV?$codecvt@_WDU_Mbstatet@@@std@@", codecvt_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(codecvt_short, 0, ".?AV?$codecvt@GDU_Mbstatet@@@std@@", codecvt_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-#else
-DEFINE_RTTI_DATA(codecvt_char, 0, ".?AV?$codecvt@DDH@std@@", codecvt_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(codecvt_wchar, 0, ".?AV?$codecvt@_WDH@std@@", codecvt_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(codecvt_short, 0, ".?AV?$codecvt@GDH@std@@", codecvt_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-#endif
-DEFINE_RTTI_DATA(numpunct_char, 0, ".?AV?$numpunct@D@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(numpunct_wchar, 0, ".?AV?$numpunct@_W@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(numpunct_short, 0, ".?AV?$numpunct@G@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(num_get_char, 0, ".?AV?$num_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(num_get_wchar, 0, ".?AV?$num_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(num_get_short, 0, ".?AV?$num_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(num_put_char, 0, ".?AV?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(num_put_wchar, 0, ".?AV?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(num_put_short, 0, ".?AV?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(time_put_char, 0, ".?AV?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(time_put_wchar, 0, ".?AV?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(time_put_short, 0, ".?AV?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(time_base, 0, ".?AUtime_base@std@@", locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(time_get_char, 0, ".?AV?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@", time_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
-DEFINE_RTTI_DATA(time_get_wchar, 0, ".?AV?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@", time_base_rtti_base_descriptor, locale_facet_rtti_base_descriptor)
+DEFINE_RTTI_DATA2(codecvt_char, 0, &codecvt_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$codecvt@DDU_Mbstatet@@@std@@")
+DEFINE_RTTI_DATA2(codecvt_char16, 0, &codecvt_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$codecvt@_SDU@std@@")
+DEFINE_RTTI_DATA2(codecvt_char32, 0, &codecvt_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$codecvt@_UDU@std@@")
+DEFINE_RTTI_DATA2(codecvt_wchar, 0, &codecvt_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$codecvt@_WDU_Mbstatet@@@std@@")
+DEFINE_RTTI_DATA2(codecvt_short, 0, &codecvt_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$codecvt@GDU_Mbstatet@@@std@@")
+#else
+DEFINE_RTTI_DATA2(codecvt_char, 0, &codecvt_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$codecvt@DDH@std@@")
+DEFINE_RTTI_DATA2(codecvt_wchar, 0, &codecvt_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$codecvt@_WDH@std@@")
+DEFINE_RTTI_DATA2(codecvt_short, 0, &codecvt_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$codecvt@GDH@std@@")
+#endif
+DEFINE_RTTI_DATA1(numpunct_char, 0, &locale_facet_rtti_base_descriptor, ".?AV?$numpunct@D@std@@")
+DEFINE_RTTI_DATA1(numpunct_wchar, 0, &locale_facet_rtti_base_descriptor, ".?AV?$numpunct@_W@std@@")
+DEFINE_RTTI_DATA1(numpunct_short, 0, &locale_facet_rtti_base_descriptor, ".?AV?$numpunct@G@std@@")
+DEFINE_RTTI_DATA1(num_get_char, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(num_get_wchar, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(num_get_short, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_get@GV?$istreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(num_put_char, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(num_put_wchar, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(num_put_short, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(time_put_char, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(time_put_wchar, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(time_put_short, 0, &locale_facet_rtti_base_descriptor, ".?AV?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@")
+DEFINE_RTTI_DATA1(time_base, 0, &locale_facet_rtti_base_descriptor, ".?AUtime_base@std@@")
+DEFINE_RTTI_DATA2(time_get_char, 0, &time_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$time_get@DV?$istreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@")
+DEFINE_RTTI_DATA2(time_get_wchar, 0, &time_base_rtti_base_descriptor, &locale_facet_rtti_base_descriptor, ".?AV?$time_get@_WV?$istreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@")
 
 __ASM_BLOCK_BEGIN(locale_vtables)
     __ASM_VTABLE(_Facet_base,
@@ -13550,39 +13554,41 @@
 
 void init_locale(void *base)
 {
-    INIT_RTTI(_Facet_base, base);
-    INIT_RTTI(locale_facet, base);
-    INIT_RTTI(locale__Locimp, base);
-    INIT_RTTI(collate_char, base);
-    INIT_RTTI(collate_wchar, base);
-    INIT_RTTI(collate_short, base);
-    INIT_RTTI(ctype_base, base);
-    INIT_RTTI(ctype_char, base);
-    INIT_RTTI(ctype_wchar, base);
-    INIT_RTTI(ctype_short, base);
-    INIT_RTTI(codecvt_base, base);
-    INIT_RTTI(codecvt_char, base);
+#ifdef RTTI_USE_RVA
+    init__Facet_base_rtti(base);
+    init_locale_facet_rtti(base);
+    init_locale__Locimp_rtti(base);
+    init_collate_char_rtti(base);
+    init_collate_wchar_rtti(base);
+    init_collate_short_rtti(base);
+    init_ctype_base_rtti(base);
+    init_ctype_char_rtti(base);
+    init_ctype_wchar_rtti(base);
+    init_ctype_short_rtti(base);
+    init_codecvt_base_rtti(base);
+    init_codecvt_char_rtti(base);
 #if _MSVCP_VER >= 140
-    INIT_RTTI(codecvt_char16, base);
-    INIT_RTTI(codecvt_char32, base);
-#endif
-    INIT_RTTI(codecvt_wchar, base);
-    INIT_RTTI(codecvt_short, base);
-    INIT_RTTI(numpunct_char, base);
-    INIT_RTTI(numpunct_wchar, base);
-    INIT_RTTI(numpunct_short, base);
-    INIT_RTTI(num_get_char, base);
-    INIT_RTTI(num_get_wchar, base);
-    INIT_RTTI(num_get_short, base);
-    INIT_RTTI(num_put_char, base);
-    INIT_RTTI(num_put_wchar, base);
-    INIT_RTTI(num_put_short, base);
-    INIT_RTTI(time_put_char, base);
-    INIT_RTTI(time_put_wchar, base);
-    INIT_RTTI(time_put_short, base);
-    INIT_RTTI(time_base, base);
-    INIT_RTTI(time_get_char, base);
-    INIT_RTTI(time_get_wchar, base);
+    init_codecvt_char16_rtti(base);
+    init_codecvt_char32_rtti(base);
+#endif
+    init_codecvt_wchar_rtti(base);
+    init_codecvt_short_rtti(base);
+    init_numpunct_char_rtti(base);
+    init_numpunct_wchar_rtti(base);
+    init_numpunct_short_rtti(base);
+    init_num_get_char_rtti(base);
+    init_num_get_wchar_rtti(base);
+    init_num_get_short_rtti(base);
+    init_num_put_char_rtti(base);
+    init_num_put_wchar_rtti(base);
+    init_num_put_short_rtti(base);
+    init_time_put_char_rtti(base);
+    init_time_put_wchar_rtti(base);
+    init_time_put_short_rtti(base);
+    init_time_base_rtti(base);
+    init_time_get_char_rtti(base);
+    init_time_get_wchar_rtti(base);
+#endif
 }
 
 void free_locale(void)
