diff --git a/dlls/d3d11/view.c b/dlls/d3d11/view.c
--- a/dlls/d3d11/view.c
+++ b/dlls/d3d11/view.c
@@ -962,7 +962,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(view->device);
+        ID3D11Device2_AddRef(view->device);
         wined3d_rendertarget_view_incref(view->wined3d_view);
     }
 
@@ -978,9 +978,9 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = view->device;
+        ID3D11Device2 *device = view->device;
         wined3d_rendertarget_view_decref(view->wined3d_view);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -1110,7 +1110,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(view->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(view->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_depthstencil_view_GetPrivateData(ID3D10DepthStencilView *iface,
@@ -1307,7 +1307,7 @@
     wined3d_private_store_init(&view->private_store);
     wined3d_mutex_unlock();
     view->resource = resource;
-    ID3D11Device5_AddRef(view->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(view->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -1400,7 +1400,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(view->device);
+        ID3D11Device2_AddRef(view->device);
         wined3d_rendertarget_view_incref(view->wined3d_view);
     }
 
@@ -1416,9 +1416,9 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = view->device;
+        ID3D11Device2 *device = view->device;
         wined3d_rendertarget_view_decref(view->wined3d_view);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -1548,7 +1548,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(view->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(view->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_rendertarget_view_GetPrivateData(ID3D10RenderTargetView *iface,
@@ -1748,7 +1748,7 @@
     wined3d_private_store_init(&view->private_store);
     wined3d_mutex_unlock();
     view->resource = resource;
-    ID3D11Device5_AddRef(view->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(view->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -1842,7 +1842,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(view->device);
+        ID3D11Device2_AddRef(view->device);
         wined3d_shader_resource_view_incref(view->wined3d_view);
     }
 
@@ -1858,9 +1858,9 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = view->device;
+        ID3D11Device2 *device = view->device;
         wined3d_shader_resource_view_decref(view->wined3d_view);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -1991,7 +1991,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(view->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(view->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_shader_resource_view_GetPrivateData(ID3D10ShaderResourceView1 *iface,
@@ -2242,7 +2242,7 @@
     wined3d_private_store_init(&view->private_store);
     wined3d_mutex_unlock();
     view->resource = resource;
-    ID3D11Device5_AddRef(view->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(view->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -2321,7 +2321,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(view->device);
+        ID3D11Device2_AddRef(view->device);
         wined3d_unordered_access_view_incref(view->wined3d_view);
     }
 
@@ -2337,9 +2337,9 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = view->device;
+        ID3D11Device2 *device = view->device;
         wined3d_unordered_access_view_decref(view->wined3d_view);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -2545,7 +2545,7 @@
     wined3d_private_store_init(&view->private_store);
     wined3d_mutex_unlock();
     view->resource = resource;
-    ID3D11Device5_AddRef(view->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(view->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -2625,8 +2625,7 @@
 
     if (!refcount)
     {
-        wined3d_decoder_output_view_decref(view->wined3d_view);
-        ID3D11Device5_Release(&view->device->ID3D11Device5_iface);
+        ID3D11Device2_Release(&view->device->ID3D11Device2_iface);
         wined3d_private_store_cleanup(&view->private_store);
         free(view);
     }
@@ -2711,65 +2710,18 @@
     d3d11_video_decoder_output_view_GetDesc,
 };
 
-static void wined3d_vdov_desc_from_d3d11(struct wined3d_view_desc *wined3d_desc,
-        const D3D11_VIDEO_DECODER_OUTPUT_VIEW_DESC *desc, DXGI_FORMAT format)
-{
-    wined3d_desc->format_id = wined3dformat_from_dxgi_format(format);
-
-    wined3d_desc->flags = 0;
-    wined3d_desc->u.texture.level_idx = 0;
-    wined3d_desc->u.texture.level_count = 1;
-    wined3d_desc->u.texture.layer_idx = desc->u.Texture2D.ArraySlice;
-    wined3d_desc->u.texture.layer_count = 1;
-}
-
 static HRESULT d3d_video_decoder_output_view_init(struct d3d_video_decoder_output_view *view,
         struct d3d_device *device, ID3D11Resource *resource, const D3D11_VIDEO_DECODER_OUTPUT_VIEW_DESC *desc)
 {
-    struct wined3d_view_desc wined3d_desc;
-    D3D11_RESOURCE_DIMENSION dimension;
-    struct d3d_texture2d *texture;
-    HRESULT hr;
-
     view->ID3D11VideoDecoderOutputView_iface.lpVtbl = &d3d11_video_decoder_output_view_vtbl;
     view->refcount = 1;
-    view->desc = *desc;
-
-    if (desc->ViewDimension != D3D11_VDOV_DIMENSION_TEXTURE2D)
-    {
-        WARN("Invalid view dimension %#x.\n", desc->ViewDimension);
-        return E_INVALIDARG;
-    }
-
-    ID3D11Resource_GetType(resource, &dimension);
-    if (dimension != D3D11_RESOURCE_DIMENSION_TEXTURE2D)
-    {
-        WARN("Invalid resource dimension %#x.\n", dimension);
-        return E_INVALIDARG;
-    }
-
-    if (!(texture = unsafe_impl_from_ID3D11Texture2D((ID3D11Texture2D *)resource)))
-    {
-        ERR("Cannot get implementation from ID3D11Texture2D.\n");
-        return E_FAIL;
-    }
-
-    wined3d_vdov_desc_from_d3d11(&wined3d_desc, &view->desc, texture->desc.Format);
 
     wined3d_mutex_lock();
-    if (FAILED(hr = wined3d_decoder_output_view_create(&wined3d_desc,
-            texture->wined3d_texture, NULL, &d3d_null_wined3d_parent_ops, &view->wined3d_view)))
-    {
-        wined3d_mutex_unlock();
-        WARN("Failed to create a wined3d video decoder output view, hr %#lx.\n", hr);
-        return hr;
-    }
-
     wined3d_private_store_init(&view->private_store);
     wined3d_mutex_unlock();
     view->resource = resource;
     view->device = device;
-    ID3D11Device5_AddRef(&device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(&device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -2797,12 +2749,3 @@
 
     return S_OK;
 }
-
-struct d3d_video_decoder_output_view *unsafe_impl_from_ID3D11VideoDecoderOutputView(ID3D11VideoDecoderOutputView *iface)
-{
-    if (!iface)
-        return NULL;
-    assert(iface->lpVtbl == &d3d11_video_decoder_output_view_vtbl);
-
-    return impl_from_ID3D11VideoDecoderOutputView(iface);
-}
