diff --git a/dlls/mmdevapi/client.c b/dlls/mmdevapi/client.c
--- a/dlls/mmdevapi/client.c
+++ b/dlls/mmdevapi/client.c
@@ -366,123 +366,6 @@
     return wcsdup(name);
 }
 
-static HRESULT validate_wfx(const WAVEFORMATEX *fmt, AUDCLNT_SHAREMODE share_mode)
-{
-    BOOL exclusive = (share_mode == AUDCLNT_SHAREMODE_EXCLUSIVE);
-    const WAVEFORMATEXTENSIBLE *fmtx = (const void *)fmt;
-    HRESULT ret = S_OK;
-
-    if (share_mode != AUDCLNT_SHAREMODE_SHARED && share_mode != AUDCLNT_SHAREMODE_EXCLUSIVE)
-        return E_INVALIDARG;
-
-    if (fmt->wFormatTag == WAVE_FORMAT_EXTENSIBLE) {
-        if (fmt->cbSize < sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX))
-            ret = E_INVALIDARG;
-        else if (fmt->nAvgBytesPerSec == 0 || fmt->nBlockAlign == 0 ||
-                (fmtx->Samples.wValidBitsPerSample > fmt->wBitsPerSample))
-            ret = E_INVALIDARG;
-        else if (fmt->nChannels == 0)
-            ret = AUDCLNT_E_UNSUPPORTED_FORMAT;
-    }
-
-    if (FAILED(ret))
-        return ret;
-
-    switch (fmt->wFormatTag) {
-        case WAVE_FORMAT_EXTENSIBLE:
-            if ((fmt->cbSize != sizeof(WAVEFORMATEXTENSIBLE) - sizeof(WAVEFORMATEX) &&
-                fmt->cbSize != sizeof(WAVEFORMATEXTENSIBLE)) ||
-                fmt->nBlockAlign != fmt->wBitsPerSample / 8 * fmt->nChannels ||
-                fmtx->Samples.wValidBitsPerSample > fmt->wBitsPerSample ||
-                fmt->nAvgBytesPerSec != fmt->nBlockAlign * fmt->nSamplesPerSec) {
-                ret = E_INVALIDARG;
-            }
-
-            if (exclusive) {
-                UINT32 mask = 0, i, channels = 0;
-
-                if (!(fmtx->dwChannelMask & (SPEAKER_ALL | SPEAKER_RESERVED))) {
-                    for (i = 1; !(i & SPEAKER_RESERVED); i <<= 1) {
-                        if (i & fmtx->dwChannelMask) {
-                            mask |= i;
-                            ++channels;
-                        }
-                    }
-
-                    if (channels != fmt->nChannels || (fmtx->dwChannelMask & ~mask)) {
-                        ret = AUDCLNT_E_UNSUPPORTED_FORMAT;
-                        break;
-                    }
-                } else {
-                    ret = AUDCLNT_E_UNSUPPORTED_FORMAT;
-                    break;
-                }
-            }
-
-            if (IsEqualGUID(&fmtx->SubFormat, &KSDATAFORMAT_SUBTYPE_IEEE_FLOAT)) {
-                if (fmt->wBitsPerSample != 32) {
-                    ret = E_INVALIDARG;
-                    break;
-                }
-
-                if (fmtx->Samples.wValidBitsPerSample != fmt->wBitsPerSample) {
-                    ret = S_FALSE;
-                }
-            } else if (IsEqualGUID(&fmtx->SubFormat, &KSDATAFORMAT_SUBTYPE_PCM)) {
-                if (!fmt->wBitsPerSample || fmt->wBitsPerSample > 32 || fmt->wBitsPerSample % 8) {
-                    ret = E_INVALIDARG;
-                    break;
-                }
-
-                if (fmtx->Samples.wValidBitsPerSample != fmt->wBitsPerSample &&
-                !(fmt->wBitsPerSample == 32 &&
-                    fmtx->Samples.wValidBitsPerSample == 24)) {
-                    ret = S_FALSE;
-                    break;
-                }
-            } else {
-                ret = AUDCLNT_E_UNSUPPORTED_FORMAT;
-                break;
-            }
-            break;
-
-        case WAVE_FORMAT_ALAW:
-        case WAVE_FORMAT_MULAW:
-            if (fmt->wBitsPerSample != 8) {
-                ret = E_INVALIDARG;
-                break;
-            }
-        /* Fall-through */
-        case WAVE_FORMAT_IEEE_FLOAT:
-            if (fmt->wFormatTag == WAVE_FORMAT_IEEE_FLOAT && fmt->wBitsPerSample != 32) {
-                ret = E_INVALIDARG;
-                break;
-            }
-        /* Fall-through */
-        case WAVE_FORMAT_PCM:
-            if (fmt->wFormatTag == WAVE_FORMAT_PCM &&
-                    (!fmt->wBitsPerSample || fmt->wBitsPerSample > 32 || fmt->wBitsPerSample % 8)) {
-                ret = E_INVALIDARG;
-                break;
-            }
-
-            if (fmt->nChannels > 2) {
-                ret = AUDCLNT_E_UNSUPPORTED_FORMAT;
-                break;
-            }
-
-            /* fmt->cbSize, fmt->nBlockAlign and fmt->nAvgBytesPerSec seem to be
-            * ignored, invalid values are happily accepted. */
-            break;
-
-        default:
-            ret = AUDCLNT_E_UNSUPPORTED_FORMAT;
-            break;
-    }
-
-    return ret;
-}
-
 static HRESULT stream_init(struct audio_client *client, const BOOLEAN force_def_period,
                            const AUDCLNT_SHAREMODE mode, const DWORD flags,
                            REFERENCE_TIME duration, REFERENCE_TIME period,
@@ -492,7 +375,6 @@
     UINT32 i, channel_count;
     stream_handle stream;
     WCHAR *name;
-    HRESULT hr;
 
     if (!fmt)
         return E_POINTER;
@@ -515,15 +397,6 @@
         FIXME("Unknown flags: %08lx\n", flags);
         return E_INVALIDARG;
     }
-    if (flags & AUDCLNT_STREAMFLAGS_CROSSPROCESS)
-        FIXME("Cross-process sessions not supported\n");
-
-    hr = validate_wfx(fmt, mode);
-
-    if (hr == S_FALSE)
-        hr = AUDCLNT_E_UNSUPPORTED_FORMAT;
-    if (hr != S_OK)
-        return hr;
 
     sessions_lock();
 
@@ -884,49 +757,32 @@
 {
     struct audio_client *This = impl_from_IAudioClient3(iface);
     struct is_format_supported_params params;
-    HRESULT hr;
 
     TRACE("(%p)->(%x, %p, %p)\n", This, mode, fmt, out);
 
-    if (out)
+    if (fmt)
+        dump_fmt(fmt);
+
+    params.device  = This->device_name;
+    params.flow    = This->dataflow;
+    params.share   = mode;
+    params.fmt_in  = fmt;
+    params.fmt_out = NULL;
+
+    if (out) {
         *out = NULL;
-
-    if (!fmt || (mode == AUDCLNT_SHAREMODE_SHARED && !out))
-        return E_POINTER;
-
-    dump_fmt(fmt);
-
-    hr = validate_wfx(fmt, mode);
-
-    if (FAILED(hr))
-        return hr;
-
-    if (hr == S_OK) {
-        params.device  = This->device_name;
-        params.flow    = This->dataflow;
-        params.share   = mode;
-        params.fmt_in  = fmt;
-
-        wine_unix_call(is_format_supported, &params);
-
-        hr = params.result;
-    }
-
-    if (hr == S_FALSE) {
-        if (mode == AUDCLNT_SHAREMODE_EXCLUSIVE) {
-            return AUDCLNT_E_UNSUPPORTED_FORMAT;
-        } else {
-            if (FAILED(hr = IAudioClient3_GetMixFormat(iface, out)))
-                return hr;
-            return S_FALSE;
-        }
-    }
-
-    if (hr == AUDCLNT_E_EXCLUSIVE_MODE_NOT_ALLOWED && This->dataflow == eCapture)
-        hr = AUDCLNT_E_UNSUPPORTED_FORMAT;
-
-    return hr;
-
+        if (mode == AUDCLNT_SHAREMODE_SHARED)
+            params.fmt_out = CoTaskMemAlloc(sizeof(*params.fmt_out));
+    }
+
+    wine_unix_call(is_format_supported, &params);
+
+    if (params.result == S_FALSE)
+        *out = &params.fmt_out->Format;
+    else
+        CoTaskMemFree(params.fmt_out);
+
+    return params.result;
 }
 
 static HRESULT WINAPI client_GetMixFormat(IAudioClient3 *iface, WAVEFORMATEX **pwfx)
@@ -1208,7 +1064,8 @@
         return hr;
 
     *default_period_frames = def_period * format->nSamplesPerSec / (REFERENCE_TIME)10000000;
-    *min_period_frames     = min_period * format->nSamplesPerSec / (REFERENCE_TIME)10000000;
+    *min_period_frames     = (min_period * format->nSamplesPerSec + 10000000 - 1) / (REFERENCE_TIME)10000000;
+    *default_period_frames = max( *default_period_frames, *min_period_frames );
     *max_period_frames     = *default_period_frames;
     *unit_period_frames    = 1;
 
@@ -1738,7 +1595,7 @@
 
     *out = NULL;
 
-    if (!get_device_name_from_guid(guid, &name, &dataflow))
+    if (!drvs.pget_device_name_from_guid(guid, &name, &dataflow))
         return AUDCLNT_E_DEVICE_INVALIDATED;
 
     if (dataflow != eRender && dataflow != eCapture) {
