diff --git a/dlls/kernel32/tests/module.c b/dlls/kernel32/tests/module.c
--- a/dlls/kernel32/tests/module.c
+++ b/dlls/kernel32/tests/module.c
@@ -44,6 +44,8 @@
 static NTSTATUS (WINAPI *pLdrGetDllHandle)( LPCWSTR load_path, ULONG flags, const UNICODE_STRING *name, HMODULE *base );
 static NTSTATUS (WINAPI *pLdrGetDllHandleEx)( ULONG flags, LPCWSTR load_path, ULONG *dll_characteristics,
                                               const UNICODE_STRING *name, HMODULE *base );
+static NTSTATUS (WINAPI *pLdrGetDllPath)(LPCWSTR,ULONG,LPWSTR*,LPWSTR*);
+static NTSTATUS (WINAPI *pLdrLoadDll)(LPCWSTR,DWORD *,const UNICODE_STRING *,HMODULE*);
 static NTSTATUS (WINAPI *pLdrGetDllFullName)( HMODULE module, UNICODE_STRING *name );
 
 static BOOL (WINAPI *pIsApiSetImplemented)(LPCSTR);
@@ -519,9 +521,13 @@
         { { 1, 1, 2 }, 0, 2 },
     };
     char *p, path[MAX_PATH], buf[MAX_PATH], curdir[MAX_PATH];
-    WCHAR bufW[MAX_PATH];
+    WCHAR bufW[MAX_PATH], pathW[MAX_PATH];
     DLL_DIRECTORY_COOKIE cookies[4];
+    WCHAR *load_path, *dummy;
     unsigned int i, j, k;
+    UNICODE_STRING name;
+    DWORD load_flags;
+    NTSTATUS status;
     BOOL ret;
     HMODULE mod;
 
@@ -550,6 +556,12 @@
     ok( mod != NULL, "LoadLibrary failed err %lu\n", GetLastError() );
     FreeLibrary( mod );
 
+    RtlInitUnicodeString( &name, L"1\\winetestdll.dll" );
+    status = pLdrLoadDll( NULL, NULL, &name, &mod );
+    ok( !status, "got %#lx.\n", status );
+    ok( !!mod, "got NULL.\n" );
+    FreeLibrary( mod );
+
     SetLastError( 0xdeadbeef );
     sprintf( path, "%c:1\\winetestdll.dll", buf[0] );
     mod = LoadLibraryA( path );
@@ -563,9 +575,45 @@
         ok( !mod, "LoadLibrary succeeded\n" );
         ok( GetLastError() == ERROR_MOD_NOT_FOUND, "wrong error %lu\n", GetLastError() );
 
+        RtlInitUnicodeString( &name, L"1\\winetestdll.dll" );
+        mod = (void *)0xdeadbeef;
+        status = pLdrLoadDll( (void *)(ULONG_PTR)(LOAD_LIBRARY_SEARCH_SYSTEM32 | 1), NULL, &name, &mod );
+        ok( status == STATUS_DLL_NOT_FOUND, "got %#lx.\n", status );
+        ok( mod == (HMODULE)0xdeadbeef, "got %p.\n", mod );
+
+        load_flags = LOAD_LIBRARY_SEARCH_SYSTEM32;
+        mod = (void *)0xdeadbeef;
+        if (0)
+        {
+            /* crashes on Windows. */
+            pLdrLoadDll( NULL, (void *)(ULONG_PTR)load_flags, &name, &mod );
+        }
+        status = pLdrLoadDll( NULL, &load_flags, &name, &mod );
+        ok( !status, "got %#lx.\n", status );
+        FreeLibrary( mod );
+
+        status = pLdrGetDllPath( name.Buffer, LOAD_LIBRARY_SEARCH_SYSTEM32, &load_path, &dummy );
+        ok( !status, "got %#lx.\n", status );
+        mod = (void *)0xdeadbeef;
+        status = pLdrLoadDll( load_path, NULL, &name, &mod );
+        ok( status == STATUS_DLL_NOT_FOUND, "got %#lx.\n", status );
+        ok( mod == (HMODULE)0xdeadbeef, "got %p.\n", mod );
+        RtlReleasePath( load_path );
+
+        mod = (void *)0xdeadbeef;
+        status = pLdrLoadDll( L".", NULL, &name, &mod );
+        ok( !status, "got %#lx.\n", status );
+        FreeLibrary( mod );
+
         SetLastError( 0xdeadbeef );
         mod = LoadLibraryExA( path, 0, LOAD_LIBRARY_SEARCH_SYSTEM32 );
         ok( mod != NULL, "LoadLibrary failed err %lu\n", GetLastError() );
+        FreeLibrary( mod );
+
+        MultiByteToWideChar( CP_ACP, 0, path, -1, pathW, ARRAY_SIZE(pathW) );
+        RtlInitUnicodeString( &name, pathW );
+        status = pLdrLoadDll( (void *)(LOAD_LIBRARY_SEARCH_SYSTEM32 | 1), NULL, &name, &mod );
+        ok( !status, "got %#lx.\n", status );
         FreeLibrary( mod );
     }
 
@@ -614,6 +662,12 @@
         ok( !mod, "LoadLibrary succeeded\n" );
         ok( GetLastError() == ERROR_INVALID_PARAMETER, "wrong error %lu\n", GetLastError() );
 
+        RtlInitUnicodeString( &name, L"winetestdll.dll" );
+        mod = (void *)0xdeadbeef;
+        status = pLdrLoadDll( (void *)(LOAD_LIBRARY_SEARCH_SYSTEM32 | LOAD_WITH_ALTERED_SEARCH_PATH | 1), NULL, &name, &mod );
+        ok( status == STATUS_INVALID_PARAMETER, "got %#lx.\n", status );
+        ok( mod == (HMODULE)0xdeadbeef, "got %p.\n", mod );
+
         SetLastError( 0xdeadbeef );
         mod = LoadLibraryExA( "winetestdll.dll", 0, LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR | LOAD_WITH_ALTERED_SEARCH_PATH );
         ok( !mod, "LoadLibrary succeeded\n" );
@@ -672,6 +726,12 @@
     mod = LoadLibraryExA( "foo\\winetestdll.dll", 0, LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR );
     ok( !mod, "LoadLibrary succeeded\n" );
     ok( GetLastError() == ERROR_INVALID_PARAMETER, "wrong error %lu\n", GetLastError() );
+
+    RtlInitUnicodeString( &name, L"foo\\winetestdll.dll" );
+    mod = (void *)0xdeadbeef;
+    status = pLdrLoadDll( (void *)(LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR | 1), NULL, &name, &mod );
+    ok( status == STATUS_INVALID_PARAMETER, "got %#lx.\n", status );
+    ok( mod == (HMODULE)0xdeadbeef, "got %p.\n", mod );
 
     SetLastError( 0xdeadbeef );
     mod = LoadLibraryExA( "\\windows\\winetestdll.dll", 0, LOAD_LIBRARY_SEARCH_DLL_LOAD_DIR );
@@ -713,6 +773,37 @@
             ok( !mod, "%u: LoadLibrary succeeded\n", j );
             ok( GetLastError() == ERROR_MOD_NOT_FOUND || broken(GetLastError() == ERROR_NOT_ENOUGH_MEMORY),
                 "%u: wrong error %lu\n", j, GetLastError() );
+        }
+        FreeLibrary( mod );
+
+        mod = LoadLibraryExA( "winetestdll.dll", 0, LOAD_LIBRARY_SEARCH_USER_DIRS );
+        if (tests[j].expect)
+        {
+            ok( mod != NULL, "%u: LoadLibrary failed err %lu\n", j, GetLastError() );
+            GetModuleFileNameA( mod, path, MAX_PATH );
+            sprintf( p, "\\%u\\winetestdll.dll", tests[j].expect );
+            ok( !lstrcmpiA( path, buf ), "%u: wrong module %s expected %s\n", j, path, buf );
+        }
+        else
+        {
+            ok( !mod, "%u: LoadLibrary succeeded\n", j );
+            ok( GetLastError() == ERROR_MOD_NOT_FOUND || broken(GetLastError() == ERROR_NOT_ENOUGH_MEMORY),
+                "%u: wrong error %lu\n", j, GetLastError() );
+        }
+        FreeLibrary( mod );
+
+        mod = NULL;
+        RtlInitUnicodeString( &name, L"winetestdll.dll" );
+        status = pLdrLoadDll( (void *)(ULONG_PTR)(LOAD_LIBRARY_SEARCH_USER_DIRS | 1), NULL, &name, &mod );
+        if (tests[j].expect)
+        {
+            ok( !status, "got %#lx.\n", status );
+            ok( !!mod, "got NULL.\n" );
+        }
+        else
+        {
+            ok( status == STATUS_DLL_NOT_FOUND, "got %#lx.\n", status );
+            ok( !mod, "got %p.\n", mod );
         }
         FreeLibrary( mod );
 
@@ -968,6 +1059,8 @@
     MAKEFUNC(LdrSetDllDirectory);
     MAKEFUNC(LdrGetDllHandle);
     MAKEFUNC(LdrGetDllHandleEx);
+    MAKEFUNC(LdrGetDllPath);
+    MAKEFUNC(LdrLoadDll);
     MAKEFUNC(LdrGetDllFullName);
     MAKEFUNC(RtlHashUnicodeString);
     mod = GetModuleHandleA( "kernelbase.dll" );
@@ -1708,7 +1801,10 @@
     TEB *teb = NtCurrentTeb(), *thread_teb;
     THREAD_BASIC_INFORMATION tbi;
     NTSTATUS status;
+    ULONG i, count;
     HANDLE thread;
+    SIZE_T size;
+    void **ptr;
 
     ok(!!teb->ThreadLocalStoragePointer, "got NULL.\n");
 
@@ -1727,6 +1823,26 @@
     ok(!thread_teb->ThreadLocalStoragePointer, "got %p.\n", thread_teb->ThreadLocalStoragePointer);
     ResumeThread(thread);
     WaitForSingleObject(test_tls_links_started, INFINITE);
+
+    if (!is_old_loader_struct())
+    {
+        ptr = teb->ThreadLocalStoragePointer;
+        count = (ULONG_PTR)ptr[-2];
+        size = HeapSize(GetProcessHeap(), 0, ptr - 2);
+        ok(size == (count + 2) * sizeof(void *), "got count %lu, size %Iu.\n", count, size);
+
+        for (i = 0; i < count; ++i)
+        {
+            if (!ptr[i]) continue;
+            size = HeapSize(GetProcessHeap(), 0, (void **)ptr[i] - 2);
+            ok(size && size < 100000, "got %Iu.\n", size);
+        }
+
+        ptr = thread_teb->ThreadLocalStoragePointer;
+        count = (ULONG_PTR)ptr[-2];
+        size = HeapSize(GetProcessHeap(), 0, ptr - 2);
+        ok(size == (count + 2) * sizeof(void *), "got count %lu, size %Iu.\n", count, size);
+    }
 
     ok(!!thread_teb->ThreadLocalStoragePointer, "got NULL.\n");
     ok(!teb->TlsLinks.Flink, "got %p.\n", teb->TlsLinks.Flink);
@@ -1850,39 +1966,6 @@
         }
         ok( found, "Could not find %s.\n", debugstr_w(modname) );
     }
-}
-
-static void test_dont_resolve_dll_references(void)
-{
-    char tmp_path[MAX_PATH], tmp_file[MAX_PATH];
-    LDR_DATA_TABLE_ENTRY *mod;
-    NTSTATUS status;
-    HMODULE modbase;
-    DWORD ret;
-    int ires;
-
-    ret = GetTempPathA( sizeof(tmp_path), tmp_path );
-    ok( !!ret, "GetTempPathA returned %lu (err %lu)\n", ret, GetLastError() );
-
-    ires = sprintf( tmp_file, "%swtstdrdr.dll", tmp_path );
-    ok( ires >= 0 && ires < sizeof(tmp_file), "sprintf returned %d\n", ires );
-
-    create_test_dll( tmp_file );
-
-    modbase = LoadLibraryExA( tmp_file, 0, DONT_RESOLVE_DLL_REFERENCES );
-    ok( modbase != NULL, "LoadLibrary returned %p (err %lu)\n", modbase, GetLastError() );
-
-    status = LdrFindEntryForAddress( modbase, &mod );
-    ok( !status, "LdrFindEntryForAddress returned %lx\n", status );
-
-    ok( !(mod->Flags & LDR_LOAD_IN_PROGRESS), "expected LDR_LOAD_IN_PROGRESS to be unset (Flags: %lx)\n", mod->Flags );
-    ok( !(mod->Flags & LDR_PROCESS_ATTACHED), "expected LDR_PROCESS_ATTACHED to be unset (Flags: %lx)\n", mod->Flags );
-
-    ret = FreeLibrary( modbase );
-    ok( !!ret, "FreeLibrary returned %lu\n", ret );
-
-    ret = DeleteFileA( tmp_file );
-    ok( !!ret, "DeleteFileA returned %lu\n", ret );
 }
 
 #define check_dll_path(a, b) check_dll_path_( __LINE__, a, b )
@@ -2016,6 +2099,5 @@
     test_tls_links();
     test_base_address_index_tree();
     test_hash_links();
-    test_dont_resolve_dll_references();
     test_known_dlls_load();
 }
