diff --git a/dlls/d3d11/shader.c b/dlls/d3d11/shader.c
--- a/dlls/d3d11/shader.c
+++ b/dlls/d3d11/shader.c
@@ -68,7 +68,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(shader->device);
+        ID3D11Device2_AddRef(shader->device);
         wined3d_shader_incref(shader->wined3d_shader);
     }
 
@@ -84,12 +84,12 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = shader->device;
+        ID3D11Device2 *device = shader->device;
 
         wined3d_shader_decref(shader->wined3d_shader);
         /* Release the device last, it may cause the wined3d device to be
          * destroyed. */
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -194,7 +194,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(shader->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(shader->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_vertex_shader_GetPrivateData(ID3D10VertexShader *iface,
@@ -279,7 +279,7 @@
     }
     wined3d_mutex_unlock();
 
-    ID3D11Device5_AddRef(shader->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(shader->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -360,7 +360,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(shader->device);
+        ID3D11Device2_AddRef(shader->device);
         wined3d_shader_incref(shader->wined3d_shader);
     }
 
@@ -376,11 +376,11 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = shader->device;
+        ID3D11Device2 *device = shader->device;
         wined3d_shader_decref(shader->wined3d_shader);
         /* Release the device last, it may cause the wined3d device to be
          * destroyed. */
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -476,7 +476,7 @@
     }
     wined3d_mutex_unlock();
 
-    ID3D11Device5_AddRef(shader->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(shader->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -547,7 +547,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(shader->device);
+        ID3D11Device2_AddRef(shader->device);
         wined3d_shader_incref(shader->wined3d_shader);
     }
 
@@ -563,11 +563,11 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = shader->device;
+        ID3D11Device2 *device = shader->device;
         wined3d_shader_decref(shader->wined3d_shader);
         /* Release the device last, it may cause the wined3d device to be
          * destroyed. */
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -663,7 +663,7 @@
     }
     wined3d_mutex_unlock();
 
-    ID3D11Device5_AddRef(shader->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(shader->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -744,7 +744,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(shader->device);
+        ID3D11Device2_AddRef(shader->device);
         wined3d_shader_incref(shader->wined3d_shader);
     }
 
@@ -760,11 +760,11 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = shader->device;
+        ID3D11Device2 *device = shader->device;
         wined3d_shader_decref(shader->wined3d_shader);
         /* Release the device last, it may cause the wined3d device to be
          * destroyed. */
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -869,7 +869,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(shader->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(shader->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_geometry_shader_GetPrivateData(ID3D10GeometryShader *iface,
@@ -1148,7 +1148,7 @@
     }
     wined3d_mutex_unlock();
 
-    ID3D11Device5_AddRef(shader->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(shader->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -1242,7 +1242,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(shader->device);
+        ID3D11Device2_AddRef(shader->device);
         wined3d_shader_incref(shader->wined3d_shader);
     }
 
@@ -1258,11 +1258,11 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = shader->device;
+        ID3D11Device2 *device = shader->device;
         wined3d_shader_decref(shader->wined3d_shader);
         /* Release the device last, it may cause the wined3d device to be
          * destroyed. */
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -1367,7 +1367,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(shader->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(shader->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_pixel_shader_GetPrivateData(ID3D10PixelShader *iface,
@@ -1452,7 +1452,7 @@
     }
     wined3d_mutex_unlock();
 
-    ID3D11Device5_AddRef(shader->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(shader->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -1532,7 +1532,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(shader->device);
+        ID3D11Device2_AddRef(shader->device);
         wined3d_shader_incref(shader->wined3d_shader);
     }
 
@@ -1548,11 +1548,11 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = shader->device;
+        ID3D11Device2 *device = shader->device;
         wined3d_shader_decref(shader->wined3d_shader);
         /* Release the device last, it may cause the wined3d device to be
          * destroyed. */
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -1647,7 +1647,7 @@
     }
     wined3d_mutex_unlock();
 
-    ID3D11Device5_AddRef(shader->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(shader->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
@@ -1727,12 +1727,12 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = class_linkage->device;
+        ID3D11Device2 *device = class_linkage->device;
 
         wined3d_private_store_cleanup(&class_linkage->private_store);
         free(class_linkage);
 
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -1826,7 +1826,7 @@
     object->refcount = 1;
     wined3d_private_store_init(&object->private_store);
 
-    ID3D11Device5_AddRef(object->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(object->device = &device->ID3D11Device2_iface);
 
     TRACE("Created class linkage %p.\n", object);
     *class_linkage = object;
