diff --git a/dlls/msvcr110/tests/msvcr110.c b/dlls/msvcr110/tests/msvcr110.c
--- a/dlls/msvcr110/tests/msvcr110.c
+++ b/dlls/msvcr110/tests/msvcr110.c
@@ -17,7 +17,6 @@
  */
 
 #include <errno.h>
-#include <malloc.h>
 #include <stdarg.h>
 #include <stdlib.h>
 #include <wchar.h>
@@ -42,6 +41,9 @@
     Context *ctx;
 } _Context;
 
+static char* (CDECL *p_setlocale)(int category, const char* locale);
+static size_t (CDECL *p___strncnt)(const char *str, size_t count);
+
 static unsigned int (CDECL *p_CurrentScheduler_GetNumberOfVirtualProcessors)(void);
 static unsigned int (CDECL *p__CurrentScheduler__GetNumberOfVirtualProcessors)(void);
 static unsigned int (CDECL *p_CurrentScheduler_Id)(void);
@@ -50,13 +52,26 @@
 static Context* (__cdecl *p_Context_CurrentContext)(void);
 static _Context* (__cdecl *p__Context__CurrentContext)(_Context*);
 
+static int (__cdecl *p_strcmp)(const char *, const char *);
+static int (__cdecl *p_strncmp)(const char *, const char *, size_t);
+static intptr_t (__cdecl *p__get_heap_handle)(void);
+
 #define SETNOFAIL(x,y) x = (void*)GetProcAddress(module,y)
 #define SET(x,y) do { SETNOFAIL(x,y); ok(x != NULL, "Export '%s' not found\n", y); } while(0)
 
 static BOOL init(void)
 {
-    HMODULE module = GetModuleHandleA("msvcr110.dll");
-
+    HMODULE module;
+
+    module = LoadLibraryA("msvcr110.dll");
+    if (!module)
+    {
+        win_skip("msvcr110.dll not installed\n");
+        return FALSE;
+    }
+
+    SET(p_setlocale, "setlocale");
+    SET(p___strncnt, "__strncnt");
     SET(p_CurrentScheduler_GetNumberOfVirtualProcessors, "?GetNumberOfVirtualProcessors@CurrentScheduler@Concurrency@@SAIXZ");
     SET(p__CurrentScheduler__GetNumberOfVirtualProcessors, "?_GetNumberOfVirtualProcessors@_CurrentScheduler@details@Concurrency@@SAIXZ");
     SET(p_CurrentScheduler_Id, "?Id@CurrentScheduler@Concurrency@@SAIXZ");
@@ -72,6 +87,10 @@
     {
         SET(p_Context_CurrentContext, "?CurrentContext@Context@Concurrency@@SAPAV12@XZ");
     }
+
+    SET(p_strcmp, "strcmp");
+    SET(p_strncmp, "strncmp");
+    SET(p__get_heap_handle, "_get_heap_handle");
 
     return TRUE;
 }
@@ -119,7 +138,7 @@
     };
 
     for(i=0; i<ARRAY_SIZE(names); i++) {
-        ret = setlocale(LC_ALL, names[i]);
+        ret = p_setlocale(LC_ALL, names[i]);
         ok(ret != NULL, "expected success, but got NULL\n");
         if(!strcmp(names[i], "syr-SY") && GetACP() == CP_UTF8)
         {
@@ -132,17 +151,17 @@
         }
     }
 
-    ret = setlocale(LC_ALL, "en-us.1250");
+    ret = p_setlocale(LC_ALL, "en-us.1250");
     ok(!ret, "setlocale(en-us.1250) succeeded (%s)\n", ret);
 
-    ret = setlocale(LC_ALL, "zh-Hans");
+    ret = p_setlocale(LC_ALL, "zh-Hans");
     ok((ret != NULL
         || broken(ret == NULL)), /* Vista */
         "expected success, but got NULL\n");
     if (ret)
         ok(!strcmp(ret, "zh-Hans"), "setlocale zh-Hans failed\n");
 
-    ret = setlocale(LC_ALL, "zh-Hant");
+    ret = p_setlocale(LC_ALL, "zh-Hant");
     ok((ret != NULL
         || broken(ret == NULL)), /* Vista */
         "expected success, but got NULL\n");
@@ -150,7 +169,7 @@
         ok(!strcmp(ret, "zh-Hant"), "setlocale zh-Hant failed\n");
 
     /* used to return Chinese (Simplified)_China.936 */
-    ret = setlocale(LC_ALL, "chinese");
+    ret = p_setlocale(LC_ALL, "chinese");
     ok(ret != NULL, "expected success, but got NULL\n");
     if (ret)
         ok((!strcmp(ret, "Chinese_China.936")
@@ -159,7 +178,7 @@
             "setlocale chinese failed, got %s\n", ret);
 
     /* used to return Chinese (Simplified)_China.936 */
-    ret = setlocale(LC_ALL, "Chinese_China.936");
+    ret = p_setlocale(LC_ALL, "Chinese_China.936");
     ok(ret != NULL, "expected success, but got NULL\n");
     if (ret)
         ok((!strcmp(ret, "Chinese_China.936")
@@ -168,7 +187,7 @@
             "setlocale Chinese_China.936 failed, got %s\n", ret);
 
     /* used to return Chinese (Simplified)_China.936 */
-    ret = setlocale(LC_ALL, "chinese-simplified");
+    ret = p_setlocale(LC_ALL, "chinese-simplified");
     ok(ret != NULL, "expected success, but got NULL\n");
     if (ret)
         ok((!strcmp(ret, "Chinese_China.936")
@@ -176,7 +195,7 @@
             "setlocale chinese-simplified failed, got %s\n", ret);
 
     /* used to return Chinese (Simplified)_China.936 */
-    ret = setlocale(LC_ALL, "chs");
+    ret = p_setlocale(LC_ALL, "chs");
     ok(ret != NULL, "expected success, but got NULL\n");
     if (ret)
         ok((!strcmp(ret, "Chinese_China.936")
@@ -184,7 +203,7 @@
             "setlocale chs failed, got %s\n", ret);
 
     /* used to return Chinese (Traditional)_Taiwan.950 */
-    ret = setlocale(LC_ALL, "cht");
+    ret = p_setlocale(LC_ALL, "cht");
     ok(ret != NULL, "expected success, but got NULL\n");
     if (ret)
         todo_wine ok((!strcmp(ret, "Chinese (Traditional)_Hong Kong SAR.950")
@@ -192,28 +211,28 @@
             "setlocale cht failed, got %s\n", ret);
 
     /* used to return Chinese (Traditional)_Taiwan.950 */
-    ret = setlocale(LC_ALL, "chinese-traditional");
+    ret = p_setlocale(LC_ALL, "chinese-traditional");
     ok(ret != NULL, "expected success, but got NULL\n");
     if (ret)
         todo_wine ok((!strcmp(ret, "Chinese (Traditional)_Hong Kong SAR.950")
             || broken(!strcmp(ret, "Chinese (Traditional)_Taiwan.950"))), /* Vista - 7 */
             "setlocale chinese-traditional failed, got %s\n", ret);
 
-    ret = setlocale(LC_ALL, "norwegian-nynorsk");
+    ret = p_setlocale(LC_ALL, "norwegian-nynorsk");
     ok(ret != NULL, "expected success, but got NULL\n");
     if (ret)
         ok((!strcmp(ret, "Norwegian Nynorsk_Norway.1252")
             || broken(!strcmp(ret, "Norwegian (Nynorsk)_Norway.1252"))), /* Vista - 7 */
              "setlocale norwegian-nynorsk failed, got %s\n", ret);
 
-    ret = setlocale(LC_ALL, "non");
+    ret = p_setlocale(LC_ALL, "non");
     ok(ret != NULL, "expected success, but got NULL\n");
     if (ret)
         ok((!strcmp(ret, "Norwegian Nynorsk_Norway.1252")
             || broken(!strcmp(ret, "Norwegian (Nynorsk)_Norway.1252"))), /* Vista - 7 */
              "setlocale norwegian-nynorsk failed, got %s\n", ret);
 
-    setlocale(LC_ALL, "C");
+    p_setlocale(LC_ALL, "C");
 }
 
 static void test___strncnt(void)
@@ -236,11 +255,11 @@
     size_t ret;
 
     if (0) /* crashes */
-        ret = __strncnt(NULL, 1);
+        ret = p___strncnt(NULL, 1);
 
     for (i = 0; i < ARRAY_SIZE(strncnt_tests); ++i)
     {
-        ret = __strncnt(strncnt_tests[i].str, strncnt_tests[i].size);
+        ret = p___strncnt(strncnt_tests[i].str, strncnt_tests[i].size);
         ok(ret == strncnt_tests[i].ret, "%u: unexpected return value %u.\n", i, (int)ret);
     }
 }
@@ -261,38 +280,38 @@
 
 static void test_strcmp(void)
 {
-    int ret = strcmp( "abc", "abcd" );
-    ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strcmp( "", "abc" );
-    ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strcmp( "abc", "ab\xa0" );
-    ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strcmp( "ab\xb0", "ab\xa0" );
+    int ret = p_strcmp( "abc", "abcd" );
+    ok( ret == -1, "wrong ret %d\n", ret );
+    ret = p_strcmp( "", "abc" );
+    ok( ret == -1, "wrong ret %d\n", ret );
+    ret = p_strcmp( "abc", "ab\xa0" );
+    ok( ret == -1, "wrong ret %d\n", ret );
+    ret = p_strcmp( "ab\xb0", "ab\xa0" );
     ok( ret == 1, "wrong ret %d\n", ret );
-    ret = strcmp( "ab\xc2", "ab\xc2" );
-    ok( ret == 0, "wrong ret %d\n", ret );
-
-    ret = strncmp( "abc", "abcd", 3 );
-    ok( ret == 0, "wrong ret %d\n", ret );
-    ret = strncmp( "", "abc", 3 );
-    ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strncmp( "abc", "ab\xa0", 4 );
-    ok( ret == -1, "wrong ret %d\n", ret );
-    ret = strncmp( "ab\xb0", "ab\xa0", 3 );
+    ret = p_strcmp( "ab\xc2", "ab\xc2" );
+    ok( ret == 0, "wrong ret %d\n", ret );
+
+    ret = p_strncmp( "abc", "abcd", 3 );
+    ok( ret == 0, "wrong ret %d\n", ret );
+    ret = p_strncmp( "", "abc", 3 );
+    ok( ret == -1, "wrong ret %d\n", ret );
+    ret = p_strncmp( "abc", "ab\xa0", 4 );
+    ok( ret == -1, "wrong ret %d\n", ret );
+    ret = p_strncmp( "ab\xb0", "ab\xa0", 3 );
     ok( ret == 1, "wrong ret %d\n", ret );
-    ret = strncmp( "ab\xb0", "ab\xa0", 2 );
-    ok( ret == 0, "wrong ret %d\n", ret );
-    ret = strncmp( "ab\xc2", "ab\xc2", 3 );
-    ok( ret == 0, "wrong ret %d\n", ret );
-    ret = strncmp( "abc", "abd", 0 );
-    ok( ret == 0, "wrong ret %d\n", ret );
-    ret = strncmp( "abc", "abc", 12 );
+    ret = p_strncmp( "ab\xb0", "ab\xa0", 2 );
+    ok( ret == 0, "wrong ret %d\n", ret );
+    ret = p_strncmp( "ab\xc2", "ab\xc2", 3 );
+    ok( ret == 0, "wrong ret %d\n", ret );
+    ret = p_strncmp( "abc", "abd", 0 );
+    ok( ret == 0, "wrong ret %d\n", ret );
+    ret = p_strncmp( "abc", "abc", 12 );
     ok( ret == 0, "wrong ret %d\n", ret );
 }
 
 static void test__get_heap_handle(void)
 {
-    ok((HANDLE)_get_heap_handle() == GetProcessHeap(), "Expected _get_heap_handle() to return GetProcessHeap()\n");
+    ok((HANDLE)p__get_heap_handle() == GetProcessHeap(), "Expected _get_heap_handle() to return GetProcessHeap()\n");
 }
 
 START_TEST(msvcr110)
