diff --git a/dlls/gdi32/text.c b/dlls/gdi32/text.c
--- a/dlls/gdi32/text.c
+++ b/dlls/gdi32/text.c
@@ -439,7 +439,7 @@
                 for (i = 0; i < uCount; i++)
                     lpOrder[i] = i;
             }
-            return FALSE;
+            return TRUE;
         }
     }
 
@@ -780,11 +780,10 @@
 
 static void logfont_AtoW( const LOGFONTA *fontA, LPLOGFONTW fontW )
 {
-    int len = MultiByteToWideChar( CP_ACP, 0, fontA->lfFaceName,
-                                   strnlen( fontA->lfFaceName, LF_FACESIZE ),
-                                   fontW->lfFaceName, LF_FACESIZE );
-    fontW->lfFaceName[min( len, LF_FACESIZE - 1 )] = 0;
     memcpy( fontW, fontA, sizeof(LOGFONTA) - LF_FACESIZE );
+    MultiByteToWideChar( CP_ACP, 0, fontA->lfFaceName, -1, fontW->lfFaceName,
+                         LF_FACESIZE );
+    fontW->lfFaceName[LF_FACESIZE - 1] = 0;
 }
 
 static void logfont_WtoA( const LOGFONTW *fontW, LPLOGFONTA fontA )
@@ -967,6 +966,44 @@
     if (dc_attr->emf && !EMFDC_ExtTextOut( dc_attr, x, y, flags, rect, str, count, dx ))
         return FALSE;
 
+    /* HACK: Use Microsoft Sans Serif for Thai. */
+    do
+    {
+        BOOL is_thai = FALSE;
+        int i;
+
+        for (i = 0; i < count; ++i)
+        {
+            if (str[i] >= 0x0e00 && str[i] <= 0x0e7f)
+                is_thai = TRUE;
+        }
+
+        if (is_thai)
+        {
+            const WCHAR *font_name = L"Microsoft Sans Serif";
+            HFONT old_font, new_font;
+            LOGFONTW log_font;
+
+            if (!(old_font = GetCurrentObject(hdc, OBJ_FONT))
+                    || !GetObjectW(old_font, sizeof(log_font), &log_font))
+                break;
+
+            if (wcscmp(log_font.lfFaceName, font_name) != 0)
+            {
+                wcscpy(log_font.lfFaceName, font_name);
+                if (!(new_font = CreateFontIndirectW(&log_font)))
+                    break;
+
+                SelectObject(hdc, new_font);
+                ret = ExtTextOutW(hdc, x, y, flags, rect, str, count, dx);
+                DeleteObject(new_font);
+                SelectObject(hdc, old_font);
+
+                return ret;
+            }
+        }
+    } while (0);
+
     if (!(flags & (ETO_GLYPH_INDEX | ETO_IGNORELANGUAGE)) && count > 0)
     {
         UINT bidi_flags;
@@ -975,11 +1012,8 @@
         bidi_flags = (dc_attr->text_align & TA_RTLREADING) || (flags & ETO_RTLREADING)
             ? WINE_GCPW_FORCE_RTL : WINE_GCPW_FORCE_LTR;
 
-        if (BIDI_Reorder( hdc, str, count, GCP_REORDER, bidi_flags, NULL, 0, NULL,
-                      &glyphs, &glyphs_count, NULL ))
-        {
-            dx = NULL;
-        }
+        BIDI_Reorder( hdc, str, count, GCP_REORDER, bidi_flags, NULL, 0, NULL,
+                      &glyphs, &glyphs_count, NULL );
 
         flags |= ETO_IGNORELANGUAGE;
         if (glyphs)
@@ -2605,7 +2639,44 @@
 static const char dos_string[0x40] = "This is a TrueType resource file";
 static const char FONTRES[] = {'F','O','N','T','R','E','S',':'};
 
-#pragma pack(push,2)
+#include <pshpack1.h>
+struct fontdir
+{
+    WORD   num_of_resources;
+    WORD   res_id;
+    WORD   dfVersion;
+    DWORD  dfSize;
+    CHAR   dfCopyright[60];
+    WORD   dfType;
+    WORD   dfPoints;
+    WORD   dfVertRes;
+    WORD   dfHorizRes;
+    WORD   dfAscent;
+    WORD   dfInternalLeading;
+    WORD   dfExternalLeading;
+    BYTE   dfItalic;
+    BYTE   dfUnderline;
+    BYTE   dfStrikeOut;
+    WORD   dfWeight;
+    BYTE   dfCharSet;
+    WORD   dfPixWidth;
+    WORD   dfPixHeight;
+    BYTE   dfPitchAndFamily;
+    WORD   dfAvgWidth;
+    WORD   dfMaxWidth;
+    BYTE   dfFirstChar;
+    BYTE   dfLastChar;
+    BYTE   dfDefaultChar;
+    BYTE   dfBreakChar;
+    WORD   dfWidthBytes;
+    DWORD  dfDevice;
+    DWORD  dfFace;
+    DWORD  dfReserved;
+    CHAR   szFaceName[LF_FACESIZE];
+};
+#include <poppack.h>
+
+#include <pshpack2.h>
 
 struct ne_typeinfo
 {
@@ -2634,7 +2705,7 @@
     BYTE fontdir_res_name[8];
 };
 
-#pragma pack(pop)
+#include <poppack.h>
 
 static BOOL create_fot( const WCHAR *resource, const WCHAR *font_file, const struct fontdir *fontdir )
 {
@@ -2644,7 +2715,6 @@
     BYTE *ptr, *start;
     BYTE import_name_len, res_name_len, non_res_name_len, font_file_len;
     char *font_fileA, *last_part, *ext;
-    const char *face_name;
     IMAGE_DOS_HEADER dos;
     IMAGE_OS2_HEADER ne =
     {
@@ -2684,8 +2754,7 @@
     if (ext) res_name_len = ext - last_part;
     else res_name_len = import_name_len - 1;
 
-    face_name = fontdir->szFaceName + strlen( fontdir->szFaceName ) + 1;  /* skip family name */
-    non_res_name_len = sizeof( FONTRES ) + strlen( face_name );
+    non_res_name_len = sizeof( FONTRES ) + strlen( fontdir->szFaceName );
 
     ne.ne_cbnrestab = 1 + non_res_name_len + 2 + 1; /* len + string + (WORD) ord_num + 1 byte eod */
     ne.ne_restab = ne.ne_rsrctab + sizeof(rsrc_tab);
@@ -2697,7 +2766,7 @@
     rsrc_tab.scalable_name.off = (ne.ne_nrestab + ne.ne_cbnrestab + 0xf) >> 4;
     rsrc_tab.scalable_name.len = (font_file_len + 0xf) >> 4;
     rsrc_tab.fontdir_name.off  = rsrc_tab.scalable_name.off + rsrc_tab.scalable_name.len;
-    rsrc_tab.fontdir_name.len  = (sizeof(*fontdir) + 0xf) >> 4;
+    rsrc_tab.fontdir_name.len  = (fontdir->dfSize + 0xf) >> 4;
 
     size = (rsrc_tab.fontdir_name.off + rsrc_tab.fontdir_name.len) << 4;
     start = ptr = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, size );
@@ -2726,13 +2795,13 @@
     ptr = start + ne.ne_nrestab;
     *ptr++ = non_res_name_len;
     memcpy( ptr, FONTRES, sizeof(FONTRES) );
-    memcpy( ptr + sizeof(FONTRES), face_name, strlen( face_name ));
+    memcpy( ptr + sizeof(FONTRES), fontdir->szFaceName, strlen( fontdir->szFaceName ) );
 
     ptr = start + (rsrc_tab.scalable_name.off << 4);
     memcpy( ptr, font_fileA, font_file_len );
 
     ptr = start + (rsrc_tab.fontdir_name.off << 4);
-    memcpy( ptr, fontdir, sizeof(*fontdir) );
+    memcpy( ptr, fontdir, fontdir->dfSize );
 
     file = CreateFileW( resource, GENERIC_READ | GENERIC_WRITE, 0, NULL, CREATE_NEW, FILE_ATTRIBUTE_NORMAL, NULL );
     if (file != INVALID_HANDLE_VALUE)
@@ -2754,10 +2823,12 @@
 BOOL WINAPI CreateScalableFontResourceW( DWORD hidden, const WCHAR *resource_file,
                                          const WCHAR *font_file, const WCHAR *font_path )
 {
-    WCHAR path[MAX_PATH];
-    struct fontdir fontdir;
+    WCHAR path[MAX_PATH], face_name[128];
+    struct fontdir fontdir = { 0 };
     UNICODE_STRING nt_name;
-    ULONG ret;
+    TEXTMETRICW otm;
+    UINT em_square;
+    BOOL ret;
 
     TRACE("(%ld, %s, %s, %s)\n", hidden, debugstr_w(resource_file),
           debugstr_w(font_file), debugstr_w(font_path) );
@@ -2773,11 +2844,45 @@
         if (!RtlDosPathNameToNtPathName_U( path, &nt_name, NULL, NULL )) goto done;
     }
     else if (!RtlDosPathNameToNtPathName_U( font_file, &nt_name, NULL, NULL )) goto done;
-
-    ret = NtGdiMakeFontDir( hidden, (BYTE *)&fontdir, sizeof(fontdir),
-                            nt_name.Buffer, nt_name.Length + sizeof(WCHAR) );
+    ret = __wine_get_file_outline_text_metric( nt_name.Buffer, &otm, &em_square, face_name );
     RtlFreeUnicodeString( &nt_name );
-    if (ret) return create_fot( resource_file, font_file, &fontdir );
+    if (!ret) goto done;
+    if (!(otm.tmPitchAndFamily & TMPF_TRUETYPE)) goto done;
+
+    fontdir.num_of_resources  = 1;
+    fontdir.res_id            = 0;
+    fontdir.dfVersion         = 0x200;
+    fontdir.dfSize            = sizeof(fontdir);
+    strcpy( fontdir.dfCopyright, "Wine fontdir" );
+    fontdir.dfType            = 0x4003;  /* 0x0080 set if private */
+    fontdir.dfPoints          = em_square;
+    fontdir.dfVertRes         = 72;
+    fontdir.dfHorizRes        = 72;
+    fontdir.dfAscent          = otm.tmAscent;
+    fontdir.dfInternalLeading = otm.tmInternalLeading;
+    fontdir.dfExternalLeading = otm.tmExternalLeading;
+    fontdir.dfItalic          = otm.tmItalic;
+    fontdir.dfUnderline       = otm.tmUnderlined;
+    fontdir.dfStrikeOut       = otm.tmStruckOut;
+    fontdir.dfWeight          = otm.tmWeight;
+    fontdir.dfCharSet         = otm.tmCharSet;
+    fontdir.dfPixWidth        = 0;
+    fontdir.dfPixHeight       = otm.tmHeight;
+    fontdir.dfPitchAndFamily  = otm.tmPitchAndFamily;
+    fontdir.dfAvgWidth        = otm.tmAveCharWidth;
+    fontdir.dfMaxWidth        = otm.tmMaxCharWidth;
+    fontdir.dfFirstChar       = otm.tmFirstChar;
+    fontdir.dfLastChar        = otm.tmLastChar;
+    fontdir.dfDefaultChar     = otm.tmDefaultChar;
+    fontdir.dfBreakChar       = otm.tmBreakChar;
+    fontdir.dfWidthBytes      = 0;
+    fontdir.dfDevice          = 0;
+    fontdir.dfFace            = FIELD_OFFSET( struct fontdir, szFaceName );
+    fontdir.dfReserved        = 0;
+    WideCharToMultiByte( CP_ACP, 0, face_name, -1, fontdir.szFaceName, LF_FACESIZE, NULL, NULL );
+
+    if (hidden) fontdir.dfType |= 0x80;
+    return create_fot( resource_file, font_file, &fontdir );
 
 done:
     SetLastError( ERROR_INVALID_PARAMETER );
