diff --git a/dlls/winewayland.drv/waylanddrv.h b/dlls/winewayland.drv/waylanddrv.h
--- a/dlls/winewayland.drv/waylanddrv.h
+++ b/dlls/winewayland.drv/waylanddrv.h
@@ -29,15 +29,11 @@
 #include <wayland-client.h>
 #include <xkbcommon/xkbcommon.h>
 #include <xkbcommon/xkbregistry.h>
-#include "cursor-shape-v1-client-protocol.h"
 #include "pointer-constraints-unstable-v1-client-protocol.h"
 #include "relative-pointer-unstable-v1-client-protocol.h"
-#include "text-input-unstable-v3-client-protocol.h"
 #include "viewporter-client-protocol.h"
 #include "xdg-output-unstable-v1-client-protocol.h"
 #include "xdg-shell-client-protocol.h"
-#include "wlr-data-control-unstable-v1-client-protocol.h"
-#include "xdg-toplevel-icon-v1-client-protocol.h"
 
 #include "windef.h"
 #include "winbase.h"
@@ -107,28 +103,11 @@
     struct zwp_confined_pointer_v1 *zwp_confined_pointer_v1;
     struct zwp_locked_pointer_v1 *zwp_locked_pointer_v1;
     struct zwp_relative_pointer_v1 *zwp_relative_pointer_v1;
-    struct wp_cursor_shape_device_v1 *wp_cursor_shape_device_v1;
     HWND focused_hwnd;
     HWND constraint_hwnd;
-    BOOL pending_warp;
     uint32_t enter_serial;
     uint32_t button_serial;
     struct wayland_cursor cursor;
-    double accum_x;
-    double accum_y;
-    pthread_mutex_t mutex;
-};
-
-struct wayland_text_input
-{
-    struct zwp_text_input_v3 *zwp_text_input_v3;
-    struct
-    {
-        WCHAR *string;
-        DWORD cursor_pos;
-    } preedit, current_preedit;
-    WCHAR *commit_string;
-    HWND focused_hwnd;
     pthread_mutex_t mutex;
 };
 
@@ -136,26 +115,6 @@
 {
     struct wl_seat *wl_seat;
     uint32_t global_id;
-    pthread_mutex_t mutex;
-};
-
-struct wayland_data_device
-{
-    union
-    {
-        struct
-        {
-            struct zwlr_data_control_device_v1 *zwlr_data_control_device_v1;
-            struct zwlr_data_control_source_v1 *zwlr_data_control_source_v1;
-            struct zwlr_data_control_offer_v1 *clipboard_zwlr_data_control_offer_v1;
-        };
-        struct
-        {
-            struct wl_data_device *wl_data_device;
-            struct wl_data_source *wl_data_source;
-            struct wl_data_offer *clipboard_wl_data_offer;
-        };
-    };
     pthread_mutex_t mutex;
 };
 
@@ -173,20 +132,12 @@
     struct wl_subcompositor *wl_subcompositor;
     struct zwp_pointer_constraints_v1 *zwp_pointer_constraints_v1;
     struct zwp_relative_pointer_manager_v1 *zwp_relative_pointer_manager_v1;
-    struct zwp_text_input_manager_v3 *zwp_text_input_manager_v3;
-    struct zwlr_data_control_manager_v1 *zwlr_data_control_manager_v1;
-    struct wl_data_device_manager *wl_data_device_manager;
-    struct xdg_toplevel_icon_manager_v1 *xdg_toplevel_icon_manager_v1;
-    struct wp_cursor_shape_manager_v1 *wp_cursor_shape_manager_v1;
     struct wayland_seat seat;
     struct wayland_keyboard keyboard;
     struct wayland_pointer pointer;
-    struct wayland_text_input text_input;
-    struct wayland_data_device data_device;
     struct wl_list output_list;
     /* Protects the output_list and the wayland_output.current states. */
     pthread_mutex_t output_mutex;
-    LONG input_serial;
 };
 
 struct wayland_output_mode
@@ -239,34 +190,20 @@
 
 struct wayland_client_surface
 {
-    struct client_surface client;
+    LONG ref;
+    HWND hwnd;
     HWND toplevel;
     struct wl_surface *wl_surface;
     struct wl_subsurface *wl_subsurface;
     struct wp_viewport *wp_viewport;
 };
 
-struct wayland_shm_buffer
-{
-    struct wl_list link;
-    struct wl_buffer *wl_buffer;
-    int width, height;
-    uint32_t format;
-    void *map_data;
-    size_t map_size;
-    BOOL busy;
-    LONG ref;
-    HRGN damage_region;
-};
-
 struct wayland_surface
 {
     HWND hwnd;
 
     struct wl_surface *wl_surface;
     struct wp_viewport *wp_viewport;
-    struct wayland_shm_buffer *small_icon_buffer;
-    struct wayland_shm_buffer *big_icon_buffer;
 
     enum wayland_surface_role role;
     union
@@ -275,7 +212,6 @@
         {
             struct xdg_surface *xdg_surface;
             struct xdg_toplevel *xdg_toplevel;
-            struct xdg_toplevel_icon_v1 *xdg_toplevel_icon;
         };
         struct
         {
@@ -289,6 +225,18 @@
     struct wayland_window_config window;
     int content_width, content_height;
     HCURSOR hcursor;
+};
+
+struct wayland_shm_buffer
+{
+    struct wl_list link;
+    struct wl_buffer *wl_buffer;
+    int width, height;
+    void *map_data;
+    size_t map_size;
+    BOOL busy;
+    LONG ref;
+    HRGN damage_region;
 };
 
 /**********************************************************************
@@ -329,16 +277,11 @@
                                       double surface_x, double surface_y,
                                       int *window_x, int *window_y);
 struct wayland_client_surface *wayland_client_surface_create(HWND hwnd);
+BOOL wayland_client_surface_release(struct wayland_client_surface *client);
 void wayland_client_surface_attach(struct wayland_client_surface *client, HWND toplevel);
+void wayland_client_surface_detach(struct wayland_client_surface *client);
 void wayland_surface_ensure_contents(struct wayland_surface *surface);
 void wayland_surface_set_title(struct wayland_surface *surface, LPCWSTR title);
-void wayland_surface_assign_icon(struct wayland_surface *surface);
-void wayland_surface_set_icon_buffer(struct wayland_surface *surface, UINT type, const ICONINFO *ii);
-
-static inline BOOL wayland_surface_is_toplevel(struct wayland_surface *surface)
-{
-    return surface->role == WAYLAND_SURFACE_ROLE_TOPLEVEL && surface->xdg_toplevel;
-}
 
 /**********************************************************************
  *          Wayland SHM buffer
@@ -346,8 +289,6 @@
 
 struct wayland_shm_buffer *wayland_shm_buffer_create(int width, int height,
                                                      enum wl_shm_format format);
-struct wayland_shm_buffer *wayland_shm_buffer_from_color_bitmaps(HDC hdc, HBITMAP color,
-                                                                 HBITMAP mask);
 void wayland_shm_buffer_ref(struct wayland_shm_buffer *shm_buffer);
 void wayland_shm_buffer_unref(struct wayland_shm_buffer *shm_buffer);
 
@@ -371,7 +312,6 @@
     struct window_rects rects;
     BOOL is_fullscreen;
     BOOL managed;
-    BOOL layered_attribs_set;
 };
 
 struct wayland_win_data *wayland_win_data_get(HWND hwnd);
@@ -379,7 +319,6 @@
 void wayland_win_data_release(struct wayland_win_data *data);
 
 struct wayland_client_surface *get_client_surface(HWND hwnd);
-void set_client_surface(HWND hwnd, struct wayland_client_surface *client);
 BOOL set_window_surface_contents(HWND hwnd, struct wayland_shm_buffer *shm_buffer, HRGN damage_region);
 struct wayland_shm_buffer *get_window_surface_contents(HWND hwnd);
 void ensure_window_surface_contents(HWND hwnd);
@@ -402,17 +341,11 @@
 void wayland_pointer_clear_constraint(void);
 
 /**********************************************************************
- *          Wayland text input
- */
-
-void wayland_text_input_init(void);
-void wayland_text_input_deinit(void);
-
-/**********************************************************************
- *          Wayland data device
- */
-
-void wayland_data_device_init(void);
+ *          OpenGL
+ */
+
+void wayland_destroy_gl_drawable(HWND hwnd);
+void wayland_resize_gl_drawable(HWND hwnd);
 
 /**********************************************************************
  *          Helpers
@@ -438,25 +371,19 @@
  *          USER driver functions
  */
 
-LRESULT WAYLAND_ClipboardWindowProc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam);
 BOOL WAYLAND_ClipCursor(const RECT *clip, BOOL reset);
 LRESULT WAYLAND_DesktopWindowProc(HWND hwnd, UINT msg, WPARAM wp, LPARAM lp);
 void WAYLAND_DestroyWindow(HWND hwnd);
-BOOL WAYLAND_SetIMECompositionRect(HWND hwnd, RECT rect);
 void WAYLAND_SetCursor(HWND hwnd, HCURSOR hcursor);
-BOOL WAYLAND_SetCursorPos(INT x, INT y);
-void WAYLAND_SetLayeredWindowAttributes(HWND hwnd, COLORREF key, BYTE alpha, DWORD flags);
-void WAYLAND_SetWindowIcons(HWND hwnd, HICON icon, const ICONINFO *ii, HICON icon_small, const ICONINFO *ii_small);
-void WAYLAND_SetWindowStyle(HWND hwnd, INT offset, STYLESTRUCT *style);
 void WAYLAND_SetWindowText(HWND hwnd, LPCWSTR text);
 LRESULT WAYLAND_SysCommand(HWND hwnd, WPARAM wparam, LPARAM lparam, const POINT *pos);
 UINT WAYLAND_UpdateDisplayDevices(const struct gdi_device_manager *device_manager, void *param);
 LRESULT WAYLAND_WindowMessage(HWND hwnd, UINT msg, WPARAM wp, LPARAM lp);
-void WAYLAND_WindowPosChanged(HWND hwnd, HWND insert_after, HWND owner_hint, UINT swp_flags,
+void WAYLAND_WindowPosChanged(HWND hwnd, HWND insert_after, HWND owner_hint, UINT swp_flags, BOOL fullscreen,
                               const struct window_rects *new_rects, struct window_surface *surface);
 BOOL WAYLAND_WindowPosChanging(HWND hwnd, UINT swp_flags, BOOL shaped, const struct window_rects *rects);
 BOOL WAYLAND_CreateWindowSurface(HWND hwnd, BOOL layered, const RECT *surface_rect, struct window_surface **surface);
 UINT WAYLAND_VulkanInit(UINT version, void *vulkan_handle, const struct vulkan_driver_funcs **driver_funcs);
-UINT WAYLAND_OpenGLInit(UINT version, const struct opengl_funcs *opengl_funcs, const struct opengl_driver_funcs **driver_funcs);
+struct opengl_funcs *WAYLAND_wine_get_wgl_driver(UINT version);
 
 #endif /* __WINE_WAYLANDDRV_H */
