diff --git a/dlls/kernelbase/main.c b/dlls/kernelbase/main.c
--- a/dlls/kernelbase/main.c
+++ b/dlls/kernelbase/main.c
@@ -26,10 +26,10 @@
 #include "shlwapi.h"
 #include "perflib.h"
 #include "winternl.h"
-#include "winperf.h"
 
 #include "wine/debug.h"
 #include "kernelbase.h"
+#include "wine/heap.h"
 #include "wine/list.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(kernelbase);
@@ -139,10 +139,7 @@
  */
 LONG WINAPI AppPolicyGetWindowingModel(HANDLE token, AppPolicyWindowingModel *policy)
 {
-    static int once;
-
-    if(!once++)
-        FIXME("%p, %p\n", token, policy);
+    FIXME("%p, %p\n", token, policy);
 
     if(policy)
         *policy = AppPolicyWindowingModel_ClassicDesktop;
@@ -220,8 +217,7 @@
 
     size = (sizeof(PERF_COUNTERSET_INSTANCE) + template->counterset.NumCounters * sizeof(UINT64)
             + (lstrlenW( name ) + 1) * sizeof(WCHAR) + 7) & ~7;
-    inst = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY,
-                      offsetof(struct counterset_instance, instance) + size );
+    inst = heap_alloc_zero( offsetof(struct counterset_instance, instance) + size );
     if (!inst)
     {
         SetLastError( ERROR_OUTOFMEMORY );
@@ -255,7 +251,7 @@
 
     inst = CONTAINING_RECORD(block, struct counterset_instance, instance);
     list_remove( &inst->entry );
-    HeapFree( GetProcessHeap(), 0, inst );
+    heap_free( inst );
 
     return ERROR_SUCCESS;
 }
@@ -284,17 +280,16 @@
     }
 
     size = offsetof( struct counterset_template, counter[template->NumCounters] );
-    if (!(new = HeapAlloc( GetProcessHeap(), 0, size ))) return ERROR_OUTOFMEMORY;
+    if (!(new = heap_alloc( size ))) return ERROR_OUTOFMEMORY;
 
     if (prov->counterset_count)
-        new_array = HeapReAlloc( GetProcessHeap(), 0, prov->countersets,
-                                 sizeof(*prov->countersets) * (prov->counterset_count + 1) );
+        new_array = heap_realloc( prov->countersets, sizeof(*prov->countersets) * (prov->counterset_count + 1) );
     else
-        new_array = HeapAlloc( GetProcessHeap(), 0, sizeof(*prov->countersets) );
+        new_array = heap_alloc( sizeof(*prov->countersets) );
 
     if (!new_array)
     {
-        HeapFree( GetProcessHeap(), 0, new );
+        heap_free( new );
         return ERROR_OUTOFMEMORY;
     }
     memcpy( new, template, size );
@@ -306,91 +301,30 @@
     return STATUS_SUCCESS;
 }
 
-static PERF_COUNTER_INFO* get_performance_counter_info(PERF_COUNTERSET_INSTANCE *instance, ULONG counter_id)
-{
-    unsigned int i;
+/***********************************************************************
+ *           PerfSetCounterRefValue   (KERNELBASE.@)
+ */
+ULONG WINAPI PerfSetCounterRefValue(HANDLE provider, PERF_COUNTERSET_INSTANCE *instance,
+                                    ULONG counterid, void *address)
+{
+    struct perf_provider *prov = perf_provider_from_handle( provider );
     struct counterset_template *template;
     struct counterset_instance *inst;
+    unsigned int i;
+
+    FIXME( "provider %p, instance %p, counterid %lu, address %p semi-stub.\n",
+           provider, instance, counterid, address );
+
+    if (!prov || !instance || !address) return ERROR_INVALID_PARAMETER;
 
     inst = CONTAINING_RECORD(instance, struct counterset_instance, instance);
     template = inst->template;
 
     for (i = 0; i < template->counterset.NumCounters; ++i)
-        if (template->counter[i].CounterId == counter_id) return  &template->counter[i];
-
-    return NULL;
-}
-
-/***********************************************************************
- *           PerfSetCounterRefValue   (KERNELBASE.@)
- */
-ULONG WINAPI PerfSetCounterRefValue(HANDLE provider, PERF_COUNTERSET_INSTANCE *instance,
-                                    ULONG counterid, void *address)
-{
-    struct perf_provider *prov = perf_provider_from_handle( provider );
-    PERF_COUNTER_INFO* counter;
-
-    FIXME( "provider %p, instance %p, counterid %lu, address %p semi-stub.\n",
-           provider, instance, counterid, address );
-
-    if (!prov || !instance || !address) return ERROR_INVALID_PARAMETER;
-
-    counter = get_performance_counter_info(instance, counterid);
-
-    if (counter == NULL) return ERROR_NOT_FOUND;
-    if (!(counter->Attrib & PERF_ATTRIB_BY_REFERENCE)) return ERROR_INVALID_PARAMETER;
-
-    *(void **)((BYTE *)(instance + 1) + counter->Offset) = address;
-
-    return STATUS_SUCCESS;
-}
-
-/***********************************************************************
- *           PerfSetULongCounterValue   (KERNELBASE.@)
- */
-ULONG WINAPI PerfSetULongCounterValue(HANDLE provider, PERF_COUNTERSET_INSTANCE *instance,
-                                      ULONG counterid, ULONG value)
-{
-    struct perf_provider *prov = perf_provider_from_handle( provider );
-    PERF_COUNTER_INFO* counter;
-
-    TRACE( "provider %p, instance %p, counterid %lu, address %lu semi-stub.\n",
-           provider, instance, counterid, value );
-
-    if (!prov || !instance) return ERROR_INVALID_PARAMETER;
-
-    counter = get_performance_counter_info(instance, counterid);
-
-    if (counter == NULL) return ERROR_NOT_FOUND;
-    if (counter->Attrib & PERF_ATTRIB_BY_REFERENCE) return ERROR_INVALID_PARAMETER;
-    if (counter->Type & PERF_SIZE_LARGE) return ERROR_INVALID_PARAMETER;
-
-    *(ULONG*)((BYTE *)(instance + 1) + counter->Offset) = value;
-
-    return STATUS_SUCCESS;
-}
-
-/***********************************************************************
- *           PerfSetULongLongCounterValue   (KERNELBASE.@)
- */
-ULONG WINAPI PerfSetULongLongCounterValue(HANDLE provider, PERF_COUNTERSET_INSTANCE *instance,
-                                          ULONG counterid, ULONGLONG value)
-{
-    struct perf_provider *prov = perf_provider_from_handle( provider );
-    PERF_COUNTER_INFO* counter;
-
-    TRACE( "provider %p, instance %p, counterid %lu, address %I64u semi-stub.\n",
-           provider, instance, counterid, value );
-
-    if (!prov || !instance) return ERROR_INVALID_PARAMETER;
-
-    counter = get_performance_counter_info(instance, counterid);
-
-    if (counter == NULL) return ERROR_NOT_FOUND;
-    if (counter->Attrib & PERF_ATTRIB_BY_REFERENCE) return ERROR_INVALID_PARAMETER;
-    if (!(counter->Type & PERF_SIZE_LARGE)) return ERROR_INVALID_PARAMETER;
-
-    *(ULONGLONG*)((BYTE *)(instance + 1) + counter->Offset) = value;
+        if (template->counter[i].CounterId == counterid) break;
+
+    if (i == template->counterset.NumCounters) return ERROR_NOT_FOUND;
+    *(void **)((BYTE *)&inst->instance + sizeof(PERF_COUNTERSET_INSTANCE) + template->counter[i].Offset) = address;
 
     return STATUS_SUCCESS;
 }
@@ -426,7 +360,7 @@
     if (context->MemAllocRoutine || context->MemFreeRoutine)
         FIXME("Memory allocation routine is not supported.\n");
 
-    if (!(prov = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*prov) ))) return ERROR_OUTOFMEMORY;
+    if (!(prov = heap_alloc_zero( sizeof(*prov) ))) return ERROR_OUTOFMEMORY;
     list_init( &prov->instance_list );
     memcpy( &prov->guid, guid, sizeof(prov->guid) );
     prov->callback = context->ControlCallback;
@@ -452,13 +386,13 @@
     LIST_FOR_EACH_ENTRY_SAFE(inst, next, &prov->instance_list, struct counterset_instance, entry)
     {
         list_remove( &inst->entry );
-        HeapFree( GetProcessHeap(), 0, inst );
+        heap_free( inst );
     }
 
     for (i = 0; i < prov->counterset_count; ++i)
-        HeapFree( GetProcessHeap(), 0, prov->countersets[i] );
-    HeapFree( GetProcessHeap(), 0, prov->countersets );
-    HeapFree( GetProcessHeap(), 0, prov );
+        heap_free( prov->countersets[i] );
+    heap_free( prov->countersets );
+    heap_free( prov );
     return STATUS_SUCCESS;
 }
 
@@ -533,7 +467,7 @@
         return E_FAIL;
 
     buflenW = *buflen;
-    langbufW = HeapAlloc(GetProcessHeap(), 0, sizeof(WCHAR) * buflenW);
+    langbufW = heap_alloc(sizeof(WCHAR) * buflenW);
     hr = GetAcceptLanguagesW(langbufW, &buflenW);
 
     if (hr == S_OK)
@@ -556,7 +490,7 @@
     }
     *buflen = buflenW ? convlen : 0;
 
-    HeapFree(GetProcessHeap(), 0, langbufW);
+    heap_free(langbufW);
     return hr;
 }
 
@@ -603,7 +537,7 @@
 
     mystrlen = (*buflen > 20) ? *buflen : 20 ;
     len = mystrlen * sizeof(WCHAR);
-    mystr = HeapAlloc(GetProcessHeap(), 0, len);
+    mystr = heap_alloc(len);
     mystr[0] = 0;
     RegOpenKeyExW(HKEY_CURRENT_USER, L"Software\\Microsoft\\Internet Explorer\\International",
                   0, KEY_QUERY_VALUE, &mykey);
@@ -615,7 +549,7 @@
     {
         lstrcpyW(langbuf, mystr);
         *buflen = len;
-        HeapFree(GetProcessHeap(), 0, mystr);
+        heap_free(mystr);
         return S_OK;
     }
 
@@ -625,7 +559,7 @@
     len = lstrlenW(mystr);
 
     memcpy(langbuf, mystr, min(*buflen, len + 1)*sizeof(WCHAR));
-    HeapFree(GetProcessHeap(), 0, mystr);
+    heap_free(mystr);
 
     if (*buflen > len)
     {
@@ -636,3 +570,9 @@
     *buflen = 0;
     return E_NOT_SUFFICIENT_BUFFER;
 }
+
+HRESULT WINAPI GetIntegratedDisplaySize( double *sz_inches )
+{
+    FIXME( "%p stub.\n", sz_inches );
+    return E_NOTIMPL;
+}
