diff --git a/dlls/comctl32/status.c b/dlls/comctl32/status.c
--- a/dlls/comctl32/status.c
+++ b/dlls/comctl32/status.c
@@ -41,6 +41,8 @@
 #include "winnls.h"
 #include "commctrl.h"
 #include "comctl32.h"
+#include "uxtheme.h"
+#include "vssym32.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(statusbar);
@@ -84,6 +86,8 @@
 #define VERT_BORDER 2
 #define HORZ_GAP    2
 
+static const WCHAR themeClass[] = L"Status";
+
 /* prototype */
 static void
 STATUSBAR_SetPartBounds (STATUS_INFO *infoPtr);
@@ -111,17 +115,15 @@
 }
 
 static void
-STATUSBAR_DrawSizeGrip (HWND hwnd, HDC hdc, LPRECT lpRect)
+STATUSBAR_DrawSizeGrip (HTHEME theme, HDC hdc, LPRECT lpRect)
 {
     RECT rc = *lpRect;
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme(hwnd);
+
+    TRACE("draw size grip %s\n", wine_dbgstr_rect(lpRect));
 
     if (theme)
     {
         SIZE gripperSize;
-
-        TRACE("draw size grip %s\n", wine_dbgstr_rect(lpRect));
         if (SUCCEEDED (GetThemePartSize (theme, hdc, SP_GRIPPER, 0, lpRect, 
             TS_DRAW, &gripperSize)))
         {
@@ -131,52 +133,40 @@
                 return;
         }
     }
-#endif
-
-    TRACE("draw size grip %s\n", wine_dbgstr_rect(lpRect));
+
     rc.left = max( rc.left, rc.right - GetSystemMetrics(SM_CXVSCROLL) - 1 );
     rc.top  = max( rc.top, rc.bottom - GetSystemMetrics(SM_CYHSCROLL) - 1 );
     DrawFrameControl( hdc, &rc, DFC_SCROLL, DFCS_SCROLLSIZEGRIP );
 }
 
 static void
-STATUSBAR_DrawPartBackground (const STATUS_INFO *infoPtr, HDC hdc, const STATUSWINDOWPART *part,
-                              int itemID, RECT *rect)
-{
+STATUSBAR_DrawPart (const STATUS_INFO *infoPtr, HDC hdc, const STATUSWINDOWPART *part, int itemID)
+{
+    RECT r = part->bound;
     UINT border;
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme(infoPtr->Self);
+    HTHEME theme = GetWindowTheme (infoPtr->Self);
+    int themePart = SP_PANE;
+    int x = 0;
+
+    TRACE("part bound %s\n", wine_dbgstr_rect(&r));
 
     if (theme)
     {
-        int themePart = SP_PANE;
-
-        if ((GetWindowLongW(infoPtr->Self, GWL_STYLE) & SBARS_SIZEGRIP)
-            && (infoPtr->simple || (itemID == (infoPtr->numParts - 1))))
+        if ((GetWindowLongW (infoPtr->Self, GWL_STYLE) & SBARS_SIZEGRIP)
+            && (infoPtr->simple || (itemID == (infoPtr->numParts-1))))
             themePart = SP_GRIPPERPANE;
-        DrawThemeBackground(theme, hdc, themePart, 0, rect, NULL);
-        return;
-    }
-#endif
-
-    if (part->style & SBT_POPOUT)
-        border = BDR_RAISEDOUTER;
-    else if (part->style & SBT_NOBORDERS)
-        border = 0;
+        DrawThemeBackground(theme, hdc, themePart, 0, &r, NULL);
+    }
     else
-        border = BDR_SUNKENOUTER;
-    DrawEdge(hdc, rect, border, BF_RECT | BF_ADJUST);
-}
-
-static void
-STATUSBAR_DrawPart (const STATUS_INFO *infoPtr, HDC hdc, const STATUSWINDOWPART *part, int itemID)
-{
-    RECT r = part->bound;
-    int x = 0;
-
-    TRACE("part bound %s\n", wine_dbgstr_rect(&r));
-
-    STATUSBAR_DrawPartBackground(infoPtr, hdc, part, itemID, &r);
+    {
+        if (part->style & SBT_POPOUT)
+            border = BDR_RAISEDOUTER;
+        else if (part->style & SBT_NOBORDERS)
+            border = 0;
+        else
+            border = BDR_SUNKENOUTER;
+        DrawEdge(hdc, &r, border, BF_RECT|BF_ADJUST);
+    }
 
     if (part->hIcon) {
         INT cy = r.bottom - r.top;
@@ -214,34 +204,14 @@
     STATUSBAR_DrawPart (infoPtr, hdc, part, itemID);
 }
 
-static void
-STATUSBAR_DrawBackground (const STATUS_INFO *infoPtr, HDC hdc, const RECT *rect)
-{
-    HBRUSH brush;
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme (infoPtr->Self);
-
-    if (theme)
-    {
-        DrawThemeBackground (theme, hdc, 0, 0, rect, NULL);
-        return;
-    }
-#endif
-
-    if (infoPtr->clrBk != CLR_DEFAULT)
-        brush = CreateSolidBrush (infoPtr->clrBk);
-    else
-        brush = GetSysColorBrush (COLOR_3DFACE);
-    FillRect (hdc, rect, brush);
-    if (infoPtr->clrBk != CLR_DEFAULT)
-        DeleteObject (brush);
-}
 
 static LRESULT
 STATUSBAR_Refresh (STATUS_INFO *infoPtr, HDC hdc)
 {
     RECT   rect;
+    HBRUSH hbrBk;
     HFONT  hOldFont;
+    HTHEME theme;
 
     TRACE("\n");
     if (!IsWindowVisible(infoPtr->Self))
@@ -250,7 +220,21 @@
     STATUSBAR_SetPartBounds(infoPtr);
 
     GetClientRect (infoPtr->Self, &rect);
-    STATUSBAR_DrawBackground (infoPtr, hdc, &rect);
+
+    if ((theme = GetWindowTheme (infoPtr->Self)))
+    {
+        DrawThemeBackground(theme, hdc, 0, 0, &rect, NULL);
+    }
+    else
+    {
+        if (infoPtr->clrBk != CLR_DEFAULT)
+            hbrBk = CreateSolidBrush (infoPtr->clrBk);
+        else
+            hbrBk = GetSysColorBrush (COLOR_3DFACE);
+        FillRect(hdc, &rect, hbrBk);
+        if (infoPtr->clrBk != CLR_DEFAULT)
+            DeleteObject (hbrBk);
+    }
 
     hOldFont = SelectObject (hdc, infoPtr->hFont ? infoPtr->hFont : infoPtr->hDefaultFont);
 
@@ -268,7 +252,7 @@
 
     if ((GetWindowLongW (infoPtr->Self, GWL_STYLE) & SBARS_SIZEGRIP)
             && !(GetWindowLongW (infoPtr->Notify, GWL_STYLE) & WS_MAXIMIZE))
-	    STATUSBAR_DrawSizeGrip (infoPtr->Self, hdc, &rect);
+	    STATUSBAR_DrawSizeGrip (theme, hdc, &rect);
 
     return 0;
 }
@@ -848,7 +832,7 @@
     if (infoPtr->hwndToolTip)
 	DestroyWindow (infoPtr->hwndToolTip);
 
-    COMCTL32_CloseThemeForWindow (infoPtr->Self);
+    CloseThemeData (GetWindowTheme (infoPtr->Self));
 
     SetWindowLongPtrW(infoPtr->Self, 0, 0);
     Free (infoPtr);
@@ -907,8 +891,8 @@
     infoPtr->parts[0].x = -1;
     infoPtr->parts[0].style = 0;
     infoPtr->parts[0].hIcon = 0;
-
-    COMCTL32_OpenThemeForWindow (hwnd, L"Status");
+    
+    OpenThemeData (hwnd, themeClass);
 
     if (lpCreate->lpszName && (len = lstrlenW ((LPCWSTR)lpCreate->lpszName)))
     {
@@ -1088,6 +1072,17 @@
     STATUSBAR_SetPartBounds (infoPtr);
     InvalidateRect(infoPtr->Self, NULL, FALSE);
     return TRUE;
+}
+
+
+/* update theme after a WM_THEMECHANGED message */
+static LRESULT theme_changed (const STATUS_INFO* infoPtr)
+{
+    HTHEME theme = GetWindowTheme (infoPtr->Self);
+    CloseThemeData (theme);
+    OpenThemeData (infoPtr->Self, themeClass);
+    InvalidateRect (infoPtr->Self, NULL, TRUE);
+    return 0;
 }
 
 
@@ -1216,11 +1211,6 @@
 	case WM_GETFONT:
 	    return (LRESULT)(infoPtr->hFont? infoPtr->hFont : infoPtr->hDefaultFont);
 
-        case WM_GETOBJECT:
-            if ((LONG)lParam == OBJID_QUERYCLASSNAMEIDX)
-                return 0x1000b;
-	    return DefWindowProcW (hwnd, msg, wParam, lParam);
-
 	case WM_GETTEXT:
             return STATUSBAR_WMGetText (infoPtr, (INT)wParam, (LPWSTR)lParam);
 
@@ -1275,7 +1265,7 @@
             return 0;
 
         case WM_THEMECHANGED:
-            return COMCTL32_ThemeChanged (infoPtr->Self, L"Status", TRUE, TRUE);
+            return theme_changed (infoPtr);
 
 	default:
 	    if ((msg >= WM_USER) && (msg < WM_APP) && !COMCTL32_IsReflectedMessage(msg))
@@ -1307,3 +1297,16 @@
 
     RegisterClassW (&wndClass);
 }
+
+
+/***********************************************************************
+ * STATUS_Unregister [Internal]
+ *
+ * Unregisters the status window class.
+ */
+
+void
+STATUS_Unregister (void)
+{
+    UnregisterClassW (STATUSCLASSNAMEW, NULL);
+}
