diff --git a/dlls/fusion/tests/asmenum.c b/dlls/fusion/tests/asmenum.c
--- a/dlls/fusion/tests/asmenum.c
+++ b/dlls/fusion/tests/asmenum.c
@@ -29,7 +29,6 @@
 #include "wine/test.h"
 #include "wine/list.h"
 
-static BOOL v4;
 static HRESULT (WINAPI *pCreateAssemblyEnum)(IAssemblyEnum **pEnum,
                                              IUnknown *pUnkReserved,
                                              IAssemblyName *pName,
@@ -63,21 +62,12 @@
         return FALSE;
     }
 
-    hr = pLoadLibraryShim(L"fusion.dll", L"v4.0.30319", NULL, &hfusion);
+    hr = pLoadLibraryShim(L"fusion.dll", NULL, NULL, &hfusion);
     if (FAILED(hr))
     {
-        hr = pLoadLibraryShim(L"fusion.dll", NULL, NULL, &hfusion);
-        if (FAILED(hr))
-        {
-            win_skip("fusion.dll not available\n");
-            FreeLibrary(hmscoree);
-            return FALSE;
-        }
-    }
-    else
-    {
-        v4 = TRUE;
-        trace("using .NET version 4\n");
+        win_skip("fusion.dll not available\n");
+        FreeLibrary(hmscoree);
+        return FALSE;
     }
 
     pCreateAssemblyEnum = (void *)GetProcAddress(hfusion, "CreateAssemblyEnum");
@@ -95,12 +85,12 @@
     return TRUE;
 }
 
-static inline void to_widechar(WCHAR *dest, const char *src)
+static inline void to_widechar(LPWSTR dest, LPCSTR src)
 {
     MultiByteToWideChar(CP_ACP, 0, src, -1, dest, MAX_PATH);
 }
 
-static inline void to_multibyte(char *dest, const WCHAR *src)
+static inline void to_multibyte(LPSTR dest, LPWSTR src)
 {
     WideCharToMultiByte(CP_ACP, 0, src, -1, dest, MAX_PATH, NULL, NULL);
 }
@@ -245,7 +235,6 @@
     do
     {
         char culture[MAX_PATH];
-        DWORD offset = 0;
 
         if (!strcmp(ffd.cFileName, ".") || !strcmp(ffd.cFileName, "..")) continue;
 
@@ -258,8 +247,7 @@
             if (GetFileAttributesA(path) == INVALID_FILE_ATTRIBUTES) continue;
         }
 
-        if (!strncmp(ffd.cFileName, "v4.0_", 5)) offset = 5;
-        if (!(ptr = strchr(ffd.cFileName + offset, '_'))) continue;
+        if (!(ptr = strchr(ffd.cFileName, '_'))) continue;
         *ptr++ = 0;
 
         if (*ptr != '_')
@@ -272,10 +260,10 @@
 
         ptr = strchr(ptr, '_');
         ptr++;
-        len = sizeof(format) + strlen(parent) + strlen(ffd.cFileName + offset) + strlen(culture) + strlen(ptr);
+        len = sizeof(format) + strlen(parent) + strlen(ffd.cFileName) + strlen(culture) + strlen(ptr);
 
         name = HeapAlloc(GetProcessHeap(), 0, offsetof( ASMNAME, data[len] ));
-        sprintf( name->data, format, parent, ffd.cFileName + offset, culture, ptr);
+        sprintf( name->data, format, parent, ffd.cFileName, culture, ptr);
         list_add_tail(assemblies, &name->entry);
     } while (FindNextFileA(hfind, &ffd) != 0);
 
@@ -301,30 +289,6 @@
     } while (FindNextFileA(hfind, &ffd) != 0);
 
     FindClose(hfind);
-}
-
-static void enum_gac(struct list *assemblies, const WCHAR *path)
-{
-    char buf[MAX_PATH];
-
-    to_multibyte(buf, path);
-    strcat(buf, "_32");
-    enum_gac_assemblies(assemblies, buf);
-
-    to_multibyte(buf, path);
-    strcat(buf, "_64");
-    enum_gac_assemblies(assemblies, buf);
-
-    to_multibyte(buf, path);
-    strcat(buf, "_Arm64");
-    enum_gac_assemblies(assemblies, buf);
-
-    to_multibyte(buf, path);
-    strcat(buf, "_MSIL");
-    enum_gac_assemblies(assemblies, buf);
-
-    to_multibyte(buf, path);
-    enum_gac_assemblies(assemblies, buf);
 }
 
 static void test_enumerate(void)
@@ -334,6 +298,7 @@
     IAssemblyEnum *asmenum;
     IAssemblyName *next;
     WCHAR buf[MAX_PATH];
+    CHAR path[MAX_PATH];
     CHAR disp[MAX_PATH];
     HRESULT hr;
     BOOL found;
@@ -343,8 +308,20 @@
     hr = pGetCachePath(ASM_CACHE_GAC, buf, &size);
     ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
 
-    enum_gac(&assemblies, buf);
-    if (v4) enum_gac(&assemblies, L"c:\\windows\\microsoft.net\\assembly\\gac");
+    to_multibyte(path, buf);
+    lstrcatA(path, "_32");
+    enum_gac_assemblies(&assemblies, path);
+
+    to_multibyte(path, buf);
+    lstrcatA(path, "_64");
+    enum_gac_assemblies(&assemblies, path);
+
+    to_multibyte(path, buf);
+    lstrcatA(path, "_MSIL");
+    enum_gac_assemblies(&assemblies, path);
+
+    to_multibyte(path, buf);
+    enum_gac_assemblies(&assemblies, path);
 
     asmenum = NULL;
     hr = pCreateAssemblyEnum(&asmenum, NULL, NULL, ASM_CACHE_GAC, NULL);
@@ -432,7 +409,7 @@
     lstrcatA(path, "\\Wine.dll");
     if (!create_file_data(path, path, 100))
     {
-        win_skip("Failed to open file %s (needs admin), skipping name enumeration tests\n", path);
+        win_skip("Failed to open file %s, skipping name enumeration tests\n", path);
         goto done;
     }
 
@@ -491,20 +468,24 @@
     hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
     to_multibyte(disp, buf);
     ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(!lstrcmpA(disp, exp[1]), "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[1], exp[2], disp);
-
-    IAssemblyName_Release(next);
-
-    next = NULL;
-    hr = IAssemblyEnum_GetNextAssembly(asmenum, NULL, &next, 0);
-    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(next != NULL, "Expected non-NULL next\n");
-
-    size = MAX_PATH;
-    hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
-    to_multibyte(disp, buf);
-    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(!lstrcmpA(disp, exp[2]), "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[2], exp[1], disp);
+    ok(!lstrcmpA(disp, exp[1]) ||
+       !lstrcmpA(disp, exp[2]), /* Win98 */
+       "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[1], exp[2], disp);
+
+    IAssemblyName_Release(next);
+
+    next = NULL;
+    hr = IAssemblyEnum_GetNextAssembly(asmenum, NULL, &next, 0);
+    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
+    ok(next != NULL, "Expected non-NULL next\n");
+
+    size = MAX_PATH;
+    hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
+    to_multibyte(disp, buf);
+    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
+    ok(!lstrcmpA(disp, exp[2]) ||
+       !lstrcmpA(disp, exp[1]), /* Win98 */
+       "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[2], exp[1], disp);
 
     IAssemblyName_Release(next);
 
@@ -538,20 +519,24 @@
     hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
     to_multibyte(disp, buf);
     ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(!lstrcmpA(disp, exp[4]), "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[4], exp[5], disp);
-
-    IAssemblyName_Release(next);
-
-    next = NULL;
-    hr = IAssemblyEnum_GetNextAssembly(asmenum, NULL, &next, 0);
-    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(next != NULL, "Expected non-NULL next\n");
-
-    size = MAX_PATH;
-    hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
-    to_multibyte(disp, buf);
-    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(!lstrcmpA(disp, exp[5]), "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[5], exp[4], disp);
+    ok(!lstrcmpA(disp, exp[4]) ||
+       !lstrcmpA(disp, exp[5]), /* Win98 */
+       "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[4], exp[5], disp);
+
+    IAssemblyName_Release(next);
+
+    next = NULL;
+    hr = IAssemblyEnum_GetNextAssembly(asmenum, NULL, &next, 0);
+    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
+    ok(next != NULL, "Expected non-NULL next\n");
+
+    size = MAX_PATH;
+    hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
+    to_multibyte(disp, buf);
+    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
+    ok(!lstrcmpA(disp, exp[5]) ||
+       !lstrcmpA(disp, exp[4]), /* Win98 */
+       "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[5], exp[4], disp);
 
     IAssemblyName_Release(next);
 
@@ -645,20 +630,24 @@
     hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
     to_multibyte(disp, buf);
     ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(!lstrcmpA(disp, exp[1]), "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[1], exp[2], disp);
-
-    IAssemblyName_Release(next);
-
-    next = NULL;
-    hr = IAssemblyEnum_GetNextAssembly(asmenum, NULL, &next, 0);
-    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(next != NULL, "Expected non-NULL next\n");
-
-    size = MAX_PATH;
-    hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
-    to_multibyte(disp, buf);
-    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
-    ok(!lstrcmpA(disp, exp[2]), "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[2], exp[1], disp);
+    ok(!lstrcmpA(disp, exp[1]) ||
+       !lstrcmpA(disp, exp[2]), /* Win98 */
+       "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[1], exp[2], disp);
+
+    IAssemblyName_Release(next);
+
+    next = NULL;
+    hr = IAssemblyEnum_GetNextAssembly(asmenum, NULL, &next, 0);
+    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
+    ok(next != NULL, "Expected non-NULL next\n");
+
+    size = MAX_PATH;
+    hr = IAssemblyName_GetDisplayName(next, buf, &size, 0);
+    to_multibyte(disp, buf);
+    ok(hr == S_OK, "Expected S_OK, got %08lx\n", hr);
+    ok(!lstrcmpA(disp, exp[2]) ||
+       !lstrcmpA(disp, exp[1]), /* Win98 */
+       "Expected \"%s\" or \"%s\", got \"%s\"\n", exp[2], exp[1], disp);
 
     IAssemblyName_Release(next);
 
