diff --git a/dlls/comctl32/tab.c b/dlls/comctl32/tab.c
--- a/dlls/comctl32/tab.c
+++ b/dlls/comctl32/tab.c
@@ -63,6 +63,8 @@
 #include "winnls.h"
 #include "commctrl.h"
 #include "comctl32.h"
+#include "uxtheme.h"
+#include "vssym32.h"
 #include "wine/debug.h"
 #include <math.h>
 
@@ -148,6 +150,8 @@
 #define TAB_HOTTRACK_TIMER            1
 #define TAB_HOTTRACK_TIMER_INTERVAL   100   /* milliseconds */
 
+static const WCHAR themeClass[] = L"Tab";
+
 static inline TAB_ITEM* TAB_GetItem(const TAB_INFO *infoPtr, INT i)
 {
     assert(i >= 0 && i < infoPtr->uNumItem);
@@ -742,7 +746,7 @@
 {
     if (tabIndex == -1) return;
 
-    if (COMCTL32_IsThemed(infoPtr->hwnd))
+    if (GetWindowTheme (infoPtr->hwnd))
     {
         RECT rect;
         TAB_InternalGetItemRect(infoPtr, tabIndex, &rect, NULL);
@@ -832,7 +836,7 @@
   if (out_redrawEnter != NULL)
     *out_redrawEnter = -1;
 
-  if ((infoPtr->dwStyle & TCS_HOTTRACK) || COMCTL32_IsThemed(infoPtr->hwnd))
+  if ((infoPtr->dwStyle & TCS_HOTTRACK) || GetWindowTheme(infoPtr->hwnd))
   {
     POINT pt;
     UINT  flags;
@@ -1516,7 +1520,7 @@
     else /* !TCS_BUTTONS */
     {
         InflateRect(&rTemp, -2, -2);
-        if (!COMCTL32_IsThemed(infoPtr->hwnd))
+        if (!GetWindowTheme (infoPtr->hwnd))
 	    FillRect(hdc, &rTemp, hbr);
     }
 
@@ -1688,7 +1692,7 @@
    * Setup for text output
   */
   oldBkMode = SetBkMode(hdc, TRANSPARENT);
-  if (!COMCTL32_IsThemed(infoPtr->hwnd) || (infoPtr->dwStyle & TCS_BUTTONS))
+  if (!GetWindowTheme (infoPtr->hwnd) || (infoPtr->dwStyle & TCS_BUTTONS))
   {
     if ((infoPtr->dwStyle & TCS_HOTTRACK) && (iItem == infoPtr->iHotTracked) &&
         !(infoPtr->dwStyle & TCS_FLATBUTTONS))
@@ -1945,54 +1949,6 @@
   DeleteObject( htextPen );
 }
 
-static void TAB_DrawItemThemeBackground(const TAB_INFO *infoPtr, HDC hdc, INT iItem,
-                                        const RECT *selectedRect, RECT *contentRect)
-{
-#if __WINE_COMCTL32_VERSION == 6
-    static const int partIds[8] =
-    {
-        /* Normal item */
-        TABP_TABITEM,
-        TABP_TABITEMLEFTEDGE,
-        TABP_TABITEMRIGHTEDGE,
-        TABP_TABITEMBOTHEDGE,
-        /* Selected tab */
-        TABP_TOPTABITEM,
-        TABP_TOPTABITEMLEFTEDGE,
-        TABP_TOPTABITEMRIGHTEDGE,
-        TABP_TOPTABITEMBOTHEDGE,
-    };
-    HTHEME theme = GetWindowTheme(infoPtr->hwnd);
-    int partIndex = 0, stateId = TIS_NORMAL;
-    RECT rect;
-
-    /* selected and unselected tabs have different parts */
-    if (iItem == infoPtr->iSelected)
-        partIndex += 4;
-    /* The part also differs on the position of a tab on a line.
-     * "Visually" determining the position works well enough. */
-    GetClientRect(infoPtr->hwnd, &rect);
-    if (selectedRect->left == 0)
-        partIndex += 1;
-    if (selectedRect->right == rect.right)
-        partIndex += 2;
-
-    if (iItem == infoPtr->iSelected)
-        stateId = TIS_SELECTED;
-    else if (iItem == infoPtr->iHotTracked)
-        stateId = TIS_HOT;
-    else if (iItem == infoPtr->uFocus)
-        stateId = TIS_FOCUSED;
-
-    /* Adjust rectangle for bottommost row */
-    if (TAB_GetItem(infoPtr, iItem)->rect.top == infoPtr->uNumRows - 1)
-        contentRect->bottom += 3;
-
-    DrawThemeBackground(theme, hdc, partIds[partIndex], stateId, contentRect, NULL);
-    GetThemeBackgroundContentRect(theme, hdc, partIds[partIndex], stateId, contentRect, contentRect);
-#endif /* __WINE_COMCTL32_VERSION == 6 */
-}
-
 /******************************************************************************
  * TAB_DrawItem
  *
@@ -2007,6 +1963,7 @@
   INT       clRight = 0;
   INT       clBottom = 0;
   COLORREF  bkgnd, corner;
+  HTHEME    theme;
 
   /*
    * Get the rectangle for the item.
@@ -2092,9 +2049,48 @@
        * Windows draws even side or bottom tabs themed, with wacky results.
        * However, since in Wine apps may get themed that did not opt in via
        * a manifest avoid theming when we know the result will be wrong */
-      if (COMCTL32_IsThemed(infoPtr->hwnd) && ((infoPtr->dwStyle & (TCS_VERTICAL | TCS_BOTTOM)) == 0))
+      if ((theme = GetWindowTheme (infoPtr->hwnd)) 
+          && ((infoPtr->dwStyle & (TCS_VERTICAL | TCS_BOTTOM)) == 0))
       {
-          TAB_DrawItemThemeBackground(infoPtr, hdc, iItem, &selectedRect, &r);
+          static const int partIds[8] = {
+              /* Normal item */
+              TABP_TABITEM,
+              TABP_TABITEMLEFTEDGE,
+              TABP_TABITEMRIGHTEDGE,
+              TABP_TABITEMBOTHEDGE,
+              /* Selected tab */
+              TABP_TOPTABITEM,
+              TABP_TOPTABITEMLEFTEDGE,
+              TABP_TOPTABITEMRIGHTEDGE,
+              TABP_TOPTABITEMBOTHEDGE,
+          };
+          int partIndex = 0;
+          int stateId = TIS_NORMAL;
+
+          /* selected and unselected tabs have different parts */
+          if (iItem == infoPtr->iSelected)
+              partIndex += 4;
+          /* The part also differs on the position of a tab on a line.
+           * "Visually" determining the position works well enough. */
+          GetClientRect(infoPtr->hwnd, &r1);
+          if(selectedRect.left == 0)
+              partIndex += 1;
+          if(selectedRect.right == r1.right)
+              partIndex += 2;
+
+          if (iItem == infoPtr->iSelected)
+              stateId = TIS_SELECTED;
+          else if (iItem == infoPtr->iHotTracked)
+              stateId = TIS_HOT;
+          else if (iItem == infoPtr->uFocus)
+              stateId = TIS_FOCUSED;
+
+          /* Adjust rectangle for bottommost row */
+          if (TAB_GetItem(infoPtr, iItem)->rect.top == infoPtr->uNumRows-1)
+            r.bottom += 3;
+
+          DrawThemeBackground (theme, hdc, partIds[partIndex], stateId, &r, NULL);
+          GetThemeBackgroundContentRect (theme, hdc, partIds[partIndex], stateId, &r, &r);
       }
       else if(infoPtr->dwStyle & TCS_VERTICAL)
       {
@@ -2299,22 +2295,6 @@
   }
 }
 
-static void TAB_DrawBorderBackground(const TAB_INFO *infoPtr, HDC hdc, RECT *rect)
-{
-#if __WINE_COMCTL32_VERSION == 6
-  HTHEME theme = GetWindowTheme(infoPtr->hwnd);
-
-  if (theme)
-  {
-      DrawThemeParentBackground(infoPtr->hwnd, hdc, rect);
-      DrawThemeBackground(theme, hdc, TABP_PANE, 0, rect, NULL);
-      return;
-  }
-#endif
-
-  DrawEdge(hdc, rect, EDGE_RAISED, BF_SOFT | BF_RECT);
-}
-
 /******************************************************************************
  * TAB_DrawBorder
  *
@@ -2324,6 +2304,7 @@
 static void TAB_DrawBorder(const TAB_INFO *infoPtr, HDC hdc)
 {
   RECT rect;
+  HTHEME theme = GetWindowTheme (infoPtr->hwnd);
 
   GetClientRect (infoPtr->hwnd, &rect);
 
@@ -2345,7 +2326,15 @@
 
   TRACE("border=(%s)\n", wine_dbgstr_rect(&rect));
 
-  TAB_DrawBorderBackground(infoPtr, hdc, &rect);
+  if (theme)
+  {
+      DrawThemeParentBackground(infoPtr->hwnd, hdc, &rect);
+      DrawThemeBackground (theme, hdc, TABP_PANE, 0, &rect, NULL);
+  }
+  else
+  {
+      DrawEdge(hdc, &rect, EDGE_RAISED, BF_SOFT|BF_RECT);
+  }
 }
 
 /******************************************************************************
@@ -3058,8 +3047,8 @@
     }
   }
 
-  COMCTL32_OpenThemeForWindow (infoPtr->hwnd, L"Tab");
-
+  OpenThemeData (infoPtr->hwnd, themeClass);
+  
   /*
    * We need to get text information so we need a DC and we need to select
    * a font.
@@ -3120,10 +3109,20 @@
   if (infoPtr->iHotTracked >= 0)
     KillTimer(infoPtr->hwnd, TAB_HOTTRACK_TIMER);
 
-  COMCTL32_CloseThemeForWindow (infoPtr->hwnd);
+  CloseThemeData (GetWindowTheme (infoPtr->hwnd));
 
   Free (infoPtr);
   return 0;
+}
+
+/* update theme after a WM_THEMECHANGED message */
+static LRESULT theme_changed(const TAB_INFO *infoPtr)
+{
+    HTHEME theme = GetWindowTheme (infoPtr->hwnd);
+    CloseThemeData (theme);
+    OpenThemeData (infoPtr->hwnd, themeClass);
+    InvalidateRect (infoPtr->hwnd, NULL, TRUE);
+    return 0;
 }
 
 static LRESULT TAB_NCCalcSize(WPARAM wParam)
@@ -3426,7 +3425,7 @@
       return 0;
 
     case WM_THEMECHANGED:
-      return COMCTL32_ThemeChanged (infoPtr->hwnd, L"Tab", TRUE, TRUE);
+      return theme_changed (infoPtr);
 
     case WM_KILLFOCUS:
       TAB_KillFocus(infoPtr);
@@ -3468,3 +3467,10 @@
 
   RegisterClassW (&wndClass);
 }
+
+
+void
+TAB_Unregister (void)
+{
+    UnregisterClassW (WC_TABCONTROLW, NULL);
+}
