diff --git a/dlls/kernel32/tests/debugger.c b/dlls/kernel32/tests/debugger.c
--- a/dlls/kernel32/tests/debugger.c
+++ b/dlls/kernel32/tests/debugger.c
@@ -27,6 +27,7 @@
 #include <winternl.h>
 #include <winreg.h>
 #include "wine/test.h"
+#include "wine/heap.h"
 #include "wine/rbtree.h"
 
 #ifndef STATUS_DEBUGGER_INACTIVE
@@ -94,7 +95,7 @@
     ret=RegQueryValueExA(hkey, value, NULL, &saved->type, NULL, &saved->size);
     if (ret == ERROR_SUCCESS)
     {
-        saved->data=malloc(saved->size);
+        saved->data=HeapAlloc(GetProcessHeap(), 0, saved->size);
         RegQueryValueExA(hkey, value, NULL, &saved->type, saved->data, &saved->size);
     }
     return ret;
@@ -105,7 +106,7 @@
     if (saved->data)
     {
         RegSetValueExA(hkey, saved->name, 0, saved->type, saved->data, saved->size);
-        free(saved->data);
+        HeapFree(GetProcessHeap(), 0, saved->data);
     }
     else
         RegDeleteValueA(hkey, saved->name);
@@ -118,13 +119,13 @@
 
     basename=strrchr(name, '\\');
     basename=(basename ? basename+1 : name);
-    event_name = malloc(6+strlen(basename)+1);
+    event_name=HeapAlloc(GetProcessHeap(), 0, 6+strlen(basename)+1);
 
     sprintf(event_name, "start_%s", basename);
     *start_event=CreateEventA(NULL, 0,0, event_name);
     sprintf(event_name, "done_%s", basename);
     *done_event=CreateEventA(NULL, 0,0, event_name);
-    free(event_name);
+    HeapFree(GetProcessHeap(), 0, event_name);
 }
 
 static void save_blackbox(const char* logfile, void* blackbox, int size, const char *dbgtrace)
@@ -254,7 +255,7 @@
 {
     struct debuggee_thread *thread;
     if (!ctx->thread_cnt++) wine_rb_init(&ctx->threads, debuggee_thread_compare);
-    thread = malloc(sizeof(*thread));
+    thread = heap_alloc(sizeof(*thread));
     thread->tid = tid;
     thread->tag = ctx->thread_tag;
     thread->handle = NULL;
@@ -275,7 +276,7 @@
 
     wine_rb_remove(&ctx->threads, &thread->entry);
     if (thread->handle) CloseHandle(thread->handle);
-    free(thread);
+    heap_free(thread);
 }
 
 static void *get_ip(const CONTEXT *ctx)
@@ -645,13 +646,13 @@
 
     get_file_name(dbglog);
     get_events(dbglog, &start_event, &done_event);
-    cmd = malloc(strlen(argv0)+10+strlen(dbgtasks)+1+strlen(dbglog)+2+34+1);
+    cmd=HeapAlloc(GetProcessHeap(), 0, strlen(argv0)+10+strlen(dbgtasks)+1+strlen(dbglog)+2+34+1);
     sprintf(cmd, "%s debugger %s \"%s\" %%ld %%ld", argv0, dbgtasks, dbglog);
     ret=RegSetValueExA(hkey, "debugger", 0, REG_SZ, (BYTE*)cmd, strlen(cmd)+1);
     ok(ret == ERROR_SUCCESS, "unable to set AeDebug/debugger: ret=%ld\n", ret);
-    free(cmd);
-
-    cmd = malloc(strlen(argv0) + 16);
+    HeapFree(GetProcessHeap(), 0, cmd);
+
+    cmd = HeapAlloc(GetProcessHeap(), 0, strlen(argv0) + 16);
     sprintf(cmd, "%s debugger crash", argv0);
 
     trace("running %s...\n", dbgtasks);
@@ -661,7 +662,7 @@
     startup.wShowWindow = SW_SHOWNORMAL;
     ret=CreateProcessA(NULL, cmd, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &info);
     ok(ret, "CreateProcess: err=%ld\n", GetLastError());
-    free(cmd);
+    HeapFree(GetProcessHeap(), 0, cmd);
     CloseHandle(info.hThread);
 
     /* The process exits... */
@@ -742,7 +743,7 @@
     ret=RegSetValueExA(hkey, "auto", 0, REG_SZ, (BYTE*)"1", 2);
     ok(ret == ERROR_SUCCESS, "unable to set AeDebug/auto: ret=%ld\n", ret);
 
-    cmd = malloc(strlen(argv0)+15+1);
+    cmd=HeapAlloc(GetProcessHeap(), 0, strlen(argv0)+15+1);
     sprintf(cmd, "%s debugger crash", argv0);
 
     memset(&startup, 0, sizeof(startup));
@@ -751,7 +752,7 @@
     startup.wShowWindow = SW_SHOWNORMAL;
     ret=CreateProcessA(NULL, cmd, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &info);
     ok(ret, "CreateProcess: err=%ld\n", GetLastError());
-    free(cmd);
+    HeapFree(GetProcessHeap(), 0, cmd);
     CloseHandle(info.hThread);
 
     trace("waiting for child exit...\n");
@@ -1044,7 +1045,7 @@
     ok(!ret, "DebugActiveProcess() succeeded on own process.\n");
 
     get_file_name(blackbox_file);
-    cmd = malloc(strlen(argv[0]) + strlen(arguments) + strlen(blackbox_file) + 2 + 10);
+    cmd = HeapAlloc(GetProcessHeap(), 0, strlen(argv[0]) + strlen(arguments) + strlen(blackbox_file) + 2 + 10);
     sprintf(cmd, "%s%s%08lx \"%s\"", argv[0], arguments, pid, blackbox_file);
 
     memset(&si, 0, sizeof(si));
@@ -1052,7 +1053,7 @@
     ret = CreateProcessA(NULL, cmd, NULL, NULL, FALSE, DEBUG_PROCESS, NULL, NULL, &si, &pi);
     ok(ret, "CreateProcess failed, last error %#lx.\n", GetLastError());
 
-    free(cmd);
+    HeapFree(GetProcessHeap(), 0, cmd);
 
     ret = pCheckRemoteDebuggerPresent(pi.hProcess, &debug);
     ok(ret, "CheckRemoteDebuggerPresent failed, last error %#lx.\n", GetLastError());
@@ -1306,7 +1307,7 @@
     WaitForSingleObject(event, INFINITE);
     CloseHandle(event);
 
-    cmd = malloc(strlen(argv[0]) + strlen(arguments) + 2);
+    cmd = HeapAlloc(GetProcessHeap(), 0, strlen(argv[0]) + strlen(arguments) + 2);
     sprintf(cmd, "%s %s", argv[0], arguments);
 
     memset(&si, 0, sizeof(si));
@@ -1325,7 +1326,7 @@
     blackbox.failures = child_failures;
     save_blackbox(blackbox_file, &blackbox, sizeof(blackbox), NULL);
 
-    free(cmd);
+    HeapFree(GetProcessHeap(), 0, cmd);
 }
 
 static void test_debug_children(const char *name, DWORD flag, BOOL debug_child, BOOL pass_exception)
@@ -1348,7 +1349,7 @@
     }
 
     get_file_name(blackbox_file);
-    cmd = malloc(strlen(name) + strlen(arguments) + strlen(blackbox_file) + 5);
+    cmd = HeapAlloc(GetProcessHeap(), 0, strlen(name) + strlen(arguments) + strlen(blackbox_file) + 5);
     sprintf(cmd, "%s %s \"%s\"", name, arguments, blackbox_file);
 
     p = strrchr(blackbox_file, '\\');
@@ -1370,7 +1371,7 @@
 
     ret = CreateProcessA(NULL, cmd, NULL, NULL, FALSE, flag, NULL, NULL, &si, &pi);
     ok(ret, "CreateProcess failed, last error %ld.\n", GetLastError());
-    free(cmd);
+    HeapFree(GetProcessHeap(), 0, cmd);
     if (!flag)
     {
         WaitForSingleObject(event_init, INFINITE);
@@ -1618,14 +1619,14 @@
     event = CreateEventW(&sa, FALSE, FALSE, NULL);
     ok(event != NULL, "CreateEvent failed: %lu\n", GetLastError());
 
-    cmd = malloc(strlen(argv0) + strlen(arguments) + 16);
+    cmd = heap_alloc(strlen(argv0) + strlen(arguments) + 16);
     sprintf(cmd, "%s%s%lx %u\n", argv0, arguments, (DWORD)(DWORD_PTR)event, OP_BP ? 3 : 1);
 
     memset(&si, 0, sizeof(si));
     si.cb = sizeof(si);
     ret = CreateProcessA(NULL, cmd, NULL, NULL, TRUE, DEBUG_PROCESS, NULL, NULL, &si, &pi);
     ok(ret, "CreateProcess failed, last error %#lx.\n", GetLastError());
-    free(cmd);
+    heap_free(cmd);
 
     next_event(&ctx, WAIT_EVENT_TIMEOUT);
     ok(ctx.ev.dwDebugEventCode == CREATE_PROCESS_DEBUG_EVENT, "dwDebugEventCode = %ld\n", ctx.ev.dwDebugEventCode);
@@ -2021,8 +2022,6 @@
                "ExceptionAddress = %p\n", ctx.ev.u.Exception.ExceptionRecord.ExceptionAddress);
             exception_cnt++;
             if (event_order == 1) event_order = 2; /* exception debug event after exit thread event */
-
-            if (ctx.ev.u.Exception.ExceptionRecord.ExceptionCode == EXCEPTION_ACCESS_VIOLATION) break;
         }
 
         trace("received %u exceptions\n", exception_cnt);
@@ -2037,6 +2036,7 @@
 
         /* main thread sleeps inside ntdll waiting for the event. set breakpoint there and make sure
          * ntdll can handle that. */
+        SuspendThread(ctx.main_thread->handle);
 
         fetch_thread_context(ctx.main_thread);
         ret = ReadProcessMemory(pi.hProcess, get_ip(&ctx.main_thread->ctx), &instr, 1, NULL);
@@ -2113,13 +2113,21 @@
 #if defined(__i386__)
         ok(ctx.main_thread->ctx.Eax == 0,   "Eax = %lx\n", ctx.main_thread->ctx.Eax);
         ok(ctx.main_thread->ctx.Ebx == 102, "Ebx = %lx\n", ctx.main_thread->ctx.Ebx);
+        ok(ctx.main_thread->ctx.Ecx != 103, "Ecx = %lx\n", ctx.main_thread->ctx.Ecx);
+        ok(ctx.main_thread->ctx.Edx != 104, "Edx = %lx\n", ctx.main_thread->ctx.Edx);
         ok(ctx.main_thread->ctx.Esi == 105, "Esi = %lx\n", ctx.main_thread->ctx.Esi);
         ok(ctx.main_thread->ctx.Edi == 106, "Edi = %lx\n", ctx.main_thread->ctx.Edi);
 #elif defined(__x86_64__)
         ok(ctx.main_thread->ctx.Rax == 0,   "Rax = %I64x\n", ctx.main_thread->ctx.Rax);
         ok(ctx.main_thread->ctx.Rbx == 102, "Rbx = %I64x\n", ctx.main_thread->ctx.Rbx);
+        ok(ctx.main_thread->ctx.Rcx != 103, "Rcx = %I64x\n", ctx.main_thread->ctx.Rcx);
+        ok(ctx.main_thread->ctx.Rdx != 104, "Rdx = %I64x\n", ctx.main_thread->ctx.Rdx);
         ok(ctx.main_thread->ctx.Rsi == 105, "Rsi = %I64x\n", ctx.main_thread->ctx.Rsi);
         ok(ctx.main_thread->ctx.Rdi == 106, "Rdi = %I64x\n", ctx.main_thread->ctx.Rdi);
+        ok(ctx.main_thread->ctx.R8  != 107, "R8 = %I64x\n",  ctx.main_thread->ctx.R8);
+        ok(ctx.main_thread->ctx.R9  != 108, "R9 = %I64x\n",  ctx.main_thread->ctx.R9);
+        ok(ctx.main_thread->ctx.R10 != 109, "R10 = %I64x\n", ctx.main_thread->ctx.R10);
+        ok(ctx.main_thread->ctx.R11 != 110, "R11 = %I64x\n", ctx.main_thread->ctx.R11);
         ok(ctx.main_thread->ctx.R12 == 111, "R12 = %I64x\n", ctx.main_thread->ctx.R12);
         ok(ctx.main_thread->ctx.R13 == 112, "R13 = %I64x\n", ctx.main_thread->ctx.R13);
         ok(ctx.main_thread->ctx.R14 == 113, "R14 = %I64x\n", ctx.main_thread->ctx.R14);
@@ -2241,9 +2249,7 @@
 static DWORD WINAPI debug_and_wait(void *arg)
 {
     STARTUPINFOA si = { sizeof(si) };
-    HANDLE *handles = arg;
-    HANDLE debug = handles[0];
-    HANDLE event = handles[1];
+    HANDLE debug = *(HANDLE *)arg;
     ULONG val = 0;
     NTSTATUS status;
     BOOL ret;
@@ -2255,7 +2261,6 @@
     status = pNtSetInformationDebugObject( debug, DebugObjectKillProcessOnExitInformation,
                                            &val, sizeof(val), NULL );
     ok( !status, "NtSetInformationDebugObject failed %lx\n", status );
-    SetEvent( event );
     Sleep(INFINITE);
     ExitThread(0);
 }
@@ -2277,7 +2282,7 @@
     SECURITY_ATTRIBUTES sa = { sizeof(sa), NULL, TRUE };
     OBJECT_ATTRIBUTES attr = { sizeof(attr) };
     NTSTATUS status;
-    HANDLE event, debug, thread, handles[2];
+    HANDLE event, debug, thread;
     DWORD exit_code, tid;
     ULONG val;
     BOOL ret;
@@ -2285,7 +2290,7 @@
     event = CreateEventW(&sa, FALSE, FALSE, NULL);
     ok(event != NULL, "CreateEvent failed: %lu\n", GetLastError());
 
-    cmd = malloc(strlen(argv0) + strlen(arguments) + 16);
+    cmd = heap_alloc(strlen(argv0) + strlen(arguments) + 16);
     sprintf(cmd, "%s%s%lx\n", argv0, arguments, (DWORD)(DWORD_PTR)event);
 
     status = pNtCreateDebugObject( &debug, DEBUG_ALL_ACCESS, &attr, 0 );
@@ -2355,12 +2360,8 @@
     /* checking on forced exit */
     status = pNtCreateDebugObject( &debug, DEBUG_ALL_ACCESS, &attr, DEBUG_KILL_ON_CLOSE );
     ok( !status, "NtCreateDebugObject failed %lx\n", status );
-    handles[0] = debug;
-    handles[1] = event;
-    ResetEvent( event );
-    thread = CreateThread(NULL, 0, debug_and_wait, handles, 0, &tid);
-    status = WaitForSingleObject( event, 10000 );
-    ok( !status, "thread didn't start %lx\n", status );
+    thread = CreateThread(NULL, 0, debug_and_wait, &debug, 0, &tid);
+    Sleep( 100 );
     ok( debug != 0, "no debug port\n" );
     val = DEBUG_KILL_ON_CLOSE;
     status = pNtSetInformationDebugObject( debug, DebugObjectKillProcessOnExitInformation,
@@ -2416,7 +2417,7 @@
     CloseHandle( thread );
 
     CloseHandle( event );
-    free(cmd);
+    heap_free(cmd);
 }
 
 static void test_OutputDebugString(void)
