diff --git a/dlls/kernelbase/path.c b/dlls/kernelbase/path.c
--- a/dlls/kernelbase/path.c
+++ b/dlls/kernelbase/path.c
@@ -35,6 +35,7 @@
 #include "kernelbase.h"
 #include "wine/exception.h"
 #include "wine/debug.h"
+#include "wine/heap.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(path);
 
@@ -86,7 +87,7 @@
         DWORD len;
 
         len = MultiByteToWideChar(CP_ACP, 0, str, -1, NULL, 0);
-        ret = HeapAlloc(GetProcessHeap(), 0, len * sizeof(WCHAR));
+        ret = heap_alloc(len * sizeof(WCHAR));
         MultiByteToWideChar(CP_ACP, 0, str, -1, ret, len);
     }
 
@@ -111,12 +112,7 @@
     if (new_capacity < count)
         new_capacity = max_capacity;
 
-    if (*elements)
-        new_elements = HeapReAlloc( GetProcessHeap(), 0, *elements, new_capacity * size );
-    else
-        new_elements = HeapAlloc( GetProcessHeap(), 0, new_capacity * size );
-
-    if (!new_elements)
+    if (!(new_elements = heap_realloc( *elements, new_capacity * size )))
         return false;
 
     *elements = new_elements;
@@ -2395,8 +2391,8 @@
     pathW = heap_strdupAtoW( path );
     maskW = heap_strdupAtoW( mask );
     ret = PathMatchSpecExW( pathW, maskW, flags );
-    HeapFree( GetProcessHeap(), 0, pathW );
-    HeapFree( GetProcessHeap(), 0, maskW );
+    heap_free( pathW );
+    heap_free( maskW );
     return ret;
 }
 
@@ -3049,7 +3045,7 @@
      */
     len = 2 + lstrlenW(url) + 1;
     if (*pcchPath < len)
-        tpath = HeapAlloc(GetProcessHeap(), 0, len * sizeof(WCHAR));
+        tpath = heap_alloc(len * sizeof(WCHAR));
     else
         tpath = path;
 
@@ -3143,7 +3139,7 @@
             lstrcpyW(path, tpath);
     }
     if (tpath != path)
-        HeapFree(GetProcessHeap(), 0, tpath);
+        heap_free(tpath);
 
     TRACE("Returning (%lu) %s\n", *pcchPath, wine_dbgstr_w(path));
     return hr;
@@ -3271,7 +3267,7 @@
 
     if ((hr = UrlEscapeW(urlW.Buffer, escapedW, &lenW, flags)) == E_POINTER)
     {
-        escapedW = HeapAlloc(GetProcessHeap(), 0, lenW * sizeof(WCHAR));
+        escapedW = heap_alloc(lenW * sizeof(WCHAR));
         hr = UrlEscapeW(urlW.Buffer, escapedW, &lenW, flags);
     }
 
@@ -3291,7 +3287,7 @@
         }
     }
     if (escapedW != bufW)
-        HeapFree(GetProcessHeap(), 0, escapedW);
+        heap_free(escapedW);
     RtlFreeUnicodeString(&urlW);
     return hr;
 }
@@ -3317,7 +3313,7 @@
         FIXME("Unimplemented flags: %08lx\n", flags);
     }
 
-    dst_ptr = dst = HeapAlloc(GetProcessHeap(), 0, *escaped_len * sizeof(WCHAR));
+    dst_ptr = dst = heap_alloc(*escaped_len * sizeof(WCHAR));
     if (!dst_ptr)
         return E_OUTOFMEMORY;
 
@@ -3482,7 +3478,7 @@
     }
     *escaped_len = needed;
 
-    HeapFree(GetProcessHeap(), 0, dst_ptr);
+    heap_free(dst_ptr);
     return hr;
 }
 
@@ -3497,11 +3493,11 @@
         return E_INVALIDARG;
 
     url = heap_strdupAtoW(src_url);
-    canonical = HeapAlloc(GetProcessHeap(), 0, *canonicalized_len * sizeof(WCHAR));
+    canonical = heap_alloc(*canonicalized_len * sizeof(WCHAR));
     if (!url || !canonical)
     {
-        HeapFree(GetProcessHeap(), 0, url);
-        HeapFree(GetProcessHeap(), 0, canonical);
+        heap_free(url);
+        heap_free(canonical);
         return E_OUTOFMEMORY;
     }
 
@@ -3509,8 +3505,8 @@
     if (hr == S_OK)
         WideCharToMultiByte(CP_ACP, 0, canonical, -1, canonicalized, *canonicalized_len + 1, NULL, NULL);
 
-    HeapFree(GetProcessHeap(), 0, url);
-    HeapFree(GetProcessHeap(), 0, canonical);
+    heap_free(url);
+    heap_free(canonical);
     return hr;
 }
 
@@ -4172,8 +4168,8 @@
         }
     }
 
-    HeapFree( GetProcessHeap(), 0, rewritten.string );
-    HeapFree( GetProcessHeap(), 0, url );
+    heap_free( rewritten.string );
+    heap_free( url );
 
     if (hr == S_OK)
         TRACE("result %s\n", wine_dbgstr_w(canonicalized));
@@ -4192,7 +4188,7 @@
     if (!url || !out || !out_len)
         return E_INVALIDARG;
 
-    inW = HeapAlloc(GetProcessHeap(), 0, 2 * INTERNET_MAX_URL_LENGTH * sizeof(WCHAR));
+    inW = heap_alloc(2 * INTERNET_MAX_URL_LENGTH * sizeof(WCHAR));
     outW = inW + INTERNET_MAX_URL_LENGTH;
 
     MultiByteToWideChar(CP_ACP, 0, url, -1, inW, INTERNET_MAX_URL_LENGTH);
@@ -4201,7 +4197,7 @@
     hr = UrlApplySchemeW(inW, outW, &len, flags);
     if (hr != S_OK)
     {
-        HeapFree(GetProcessHeap(), 0, inW);
+        heap_free(inW);
         return hr;
     }
 
@@ -4217,7 +4213,7 @@
 
 cleanup:
     *out_len = len;
-    HeapFree(GetProcessHeap(), 0, inW);
+    heap_free(inW);
     return hr;
 }
 
@@ -4295,12 +4291,12 @@
         }
     }
 
-    new_url = HeapAlloc(GetProcessHeap(), 0, (lstrlenW(path) + 9) * sizeof(WCHAR)); /* "file:///" + path length + 1 */
+    new_url = heap_alloc((lstrlenW(path) + 9) * sizeof(WCHAR)); /* "file:///" + path length + 1 */
     lstrcpyW(new_url, L"file:");
     if (is_drive_spec( path )) lstrcatW(new_url, L"///");
     lstrcatW(new_url, path);
     hr = UrlEscapeW(new_url, url, url_len, URL_ESCAPE_PERCENT);
-    HeapFree(GetProcessHeap(), 0, new_url);
+    heap_free(new_url);
     return hr;
 }
 
@@ -4474,7 +4470,7 @@
     if (!url || !out || !out_len || !*out_len)
         return E_INVALIDARG;
 
-    inW = HeapAlloc(GetProcessHeap(), 0, 2 * INTERNET_MAX_URL_LENGTH * sizeof(WCHAR));
+    inW = heap_alloc(2 * INTERNET_MAX_URL_LENGTH * sizeof(WCHAR));
     outW = inW + INTERNET_MAX_URL_LENGTH;
 
     MultiByteToWideChar(CP_ACP, 0, url, -1, inW, INTERNET_MAX_URL_LENGTH);
@@ -4483,7 +4479,7 @@
     hr = UrlGetPartW(inW, outW, &len, part, flags);
     if (FAILED(hr))
     {
-        HeapFree(GetProcessHeap(), 0, inW);
+        heap_free(inW);
         return hr;
     }
 
@@ -4491,12 +4487,12 @@
     if (len2 > *out_len)
     {
         *out_len = len2;
-        HeapFree(GetProcessHeap(), 0, inW);
+        heap_free(inW);
         return E_POINTER;
     }
     len2 = WideCharToMultiByte(CP_ACP, 0, outW, len + 1, out, *out_len, NULL, NULL);
     *out_len = len2 - 1;
-    HeapFree(GetProcessHeap(), 0, inW);
+    heap_free(inW);
     if (hr == S_OK && !*out_len) hr = S_FALSE;
     return hr;
 }
@@ -4836,7 +4832,7 @@
 
     if ((hr = UrlCreateFromPathW(pathW.Buffer, urlW, &lenW, reserved)) == E_POINTER)
     {
-        urlW = HeapAlloc(GetProcessHeap(), 0, lenW * sizeof(WCHAR));
+        urlW = heap_alloc(lenW * sizeof(WCHAR));
         hr = UrlCreateFromPathW(pathW.Buffer, urlW, &lenW, reserved);
     }
 
@@ -4856,7 +4852,7 @@
         }
     }
     if (urlW != bufW)
-        HeapFree(GetProcessHeap(), 0, urlW);
+        heap_free(urlW);
     RtlFreeUnicodeString(&pathW);
     return hr;
 }
@@ -4888,7 +4884,7 @@
     if (!base || !relative || !combined_len)
         return E_INVALIDARG;
 
-    baseW = HeapAlloc(GetProcessHeap(), 0, 3 * INTERNET_MAX_URL_LENGTH * sizeof(WCHAR));
+    baseW = heap_alloc(3 * INTERNET_MAX_URL_LENGTH * sizeof(WCHAR));
     relativeW = baseW + INTERNET_MAX_URL_LENGTH;
     combinedW = relativeW + INTERNET_MAX_URL_LENGTH;
 
@@ -4900,7 +4896,7 @@
     if (hr != S_OK)
     {
         *combined_len = len;
-        HeapFree(GetProcessHeap(), 0, baseW);
+        heap_free(baseW);
         return hr;
     }
 
@@ -4908,12 +4904,12 @@
     if (len2 > *combined_len)
     {
         *combined_len = len2;
-        HeapFree(GetProcessHeap(), 0, baseW);
+        heap_free(baseW);
         return E_POINTER;
     }
     WideCharToMultiByte(CP_ACP, 0, combinedW, len+1, combined, *combined_len + 1, NULL, NULL);
     *combined_len = len2;
-    HeapFree(GetProcessHeap(), 0, baseW);
+    heap_free(baseW);
     return S_OK;
 }
 
@@ -4933,7 +4929,7 @@
     relative.cbSize = sizeof(relative);
 
     /* Get space for duplicates of the input and the output */
-    preliminary = HeapAlloc(GetProcessHeap(), 0, 3 * INTERNET_MAX_URL_LENGTH * sizeof(WCHAR));
+    preliminary = heap_alloc(3 * INTERNET_MAX_URL_LENGTH * sizeof(WCHAR));
     mbase = preliminary + INTERNET_MAX_URL_LENGTH;
     canonicalized = mbase + INTERNET_MAX_URL_LENGTH;
     *preliminary = '\0';
@@ -5184,7 +5180,7 @@
         TRACE("return-%ld len=%ld, %s\n", process_case, *combined_len, debugstr_w(combined));
     }
 
-    HeapFree(GetProcessHeap(), 0, preliminary);
+    heap_free(preliminary);
     return hr;
 }
 
