diff --git a/dlls/mscoree/metahost.c b/dlls/mscoree/metahost.c
--- a/dlls/mscoree/metahost.c
+++ b/dlls/mscoree/metahost.c
@@ -711,7 +711,7 @@
 
 #ifdef __i386__
 static const WCHAR libmono2_arch_dll[] = {'\\','b','i','n','\\','l','i','b','m','o','n','o','-','2','.','0','-','x','8','6','.','d','l','l',0};
-#elif defined(__x86_64__)
+#elif defined(__x86_64__) || defined(__aarch64__)
 static const WCHAR libmono2_arch_dll[] = {'\\','b','i','n','\\','l','i','b','m','o','n','o','-','2','.','0','-','x','8','6','_','6','4','.','d','l','l',0};
 #else
 static const WCHAR libmono2_arch_dll[] = {'\\','b','i','n','\\','l','i','b','m','o','n','o','-','2','.','0','.','d','l','l',0};
@@ -791,12 +791,17 @@
 
 static BOOL get_mono_path_dos(const WCHAR *dir, LPWSTR path)
 {
-    static const WCHAR basedir[] = L"\\wine-mono-" WINE_MONO_VERSION;
+    static const WCHAR unix_prefix[] = {'\\','\\','?','\\','u','n','i','x','\\'};
+    static const WCHAR basedir[] = L"\\wine-mono";
+    static const WCHAR basedir2[] = L"\\wine-mono-" WINE_MONO_VERSION;
     LPWSTR dos_dir;
     WCHAR mono_dll_path[MAX_PATH];
     DWORD len;
     BOOL ret;
 
+    if (memcmp(dir, unix_prefix, sizeof(unix_prefix)) == 0)
+        return FALSE;  /* No drive letter for this directory */
+
     len = lstrlenW( dir ) + lstrlenW( basedir ) + 1;
     if (!(dos_dir = malloc( len * sizeof(WCHAR) ))) return FALSE;
     lstrcpyW( dos_dir, dir );
@@ -808,45 +813,65 @@
 
     free(dos_dir);
 
+    if (!ret)
+    {
+        len = lstrlenW( dir ) + lstrlenW( basedir2 ) + 1;
+        if (!(dos_dir = malloc( len * sizeof(WCHAR) ))) return FALSE;
+        lstrcpyW( dos_dir, dir );
+        lstrcatW( dos_dir, basedir2 );
+
+        ret = find_mono_dll(dos_dir, mono_dll_path);
+        if (ret)
+            lstrcpyW(path, dos_dir);
+
+        free(dos_dir);
+    }
+
     return ret;
 }
 
-static BOOL get_mono_path_unix(const WCHAR *unix_dir, LPWSTR path)
-{
-    WCHAR buffer[MAX_PATH];
-    HANDLE dir;
-    DWORD len;
-
-    if (!get_mono_path_dos( unix_dir, buffer )) return FALSE;
-
-    /* try to get a DOS name for that directory */
-
-    dir = CreateFileW( buffer, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
-                       NULL, OPEN_EXISTING, FILE_FLAG_BACKUP_SEMANTICS, 0 );
-    if (dir == INVALID_HANDLE_VALUE) return FALSE;
-    len = GetFinalPathNameByHandleW( dir, buffer, ARRAY_SIZE(buffer), VOLUME_NAME_DOS );
-    CloseHandle( dir );
-    if (!len) return FALSE;
-    if (len > ARRAY_SIZE(buffer)) return FALSE;
-    if (buffer[5] != ':') return FALSE;
-    wcscpy( path, buffer + 4 );
-    return TRUE;
+static BOOL get_mono_path_unix(const char *unix_dir, LPWSTR path)
+{
+    static WCHAR * (CDECL *p_wine_get_dos_file_name)(const char*);
+    LPWSTR dos_dir;
+    BOOL ret;
+
+    if (!p_wine_get_dos_file_name)
+    {
+        p_wine_get_dos_file_name = (void*)GetProcAddress(GetModuleHandleA("kernel32"), "wine_get_dos_file_name");
+        if (!p_wine_get_dos_file_name)
+            return FALSE;
+    }
+
+    dos_dir = p_wine_get_dos_file_name(unix_dir);
+    if (!dos_dir)
+        return FALSE;
+
+    ret = get_mono_path_dos( dos_dir, path);
+
+    HeapFree(GetProcessHeap(), 0, dos_dir);
+    return ret;
 }
 
 static BOOL get_mono_path_datadir(LPWSTR path)
 {
+    static const WCHAR winedatadirW[] = {'W','I','N','E','D','A','T','A','D','I','R',0};
+    static const WCHAR winebuilddirW[] = {'W','I','N','E','B','U','I','L','D','D','I','R',0};
+    static const WCHAR unix_prefix[] = {'\\','?','?','\\','u','n','i','x','\\',0};
+    static const WCHAR monoW[] = {'\\','m','o','n','o',0};
+    static const WCHAR dotdotmonoW[] = {'\\','.','.','\\','m','o','n','o',0};
     const WCHAR *data_dir, *suffix;
     WCHAR *package_dir;
     BOOL ret;
 
-    if ((data_dir = _wgetenv( L"WINEDATADIR" )))
-        suffix = L"\\mono";
-    else if ((data_dir = _wgetenv( L"WINEBUILDDIR" )))
-        suffix = L"\\..\\mono";
+    if ((data_dir = _wgetenv( winedatadirW )))
+        suffix = monoW;
+    else if ((data_dir = _wgetenv( winebuilddirW )))
+        suffix = dotdotmonoW;
     else
         return FALSE;
 
-    if (!wcsncmp( data_dir, L"\\??\\unix", 8 )) return FALSE;
+    if (!wcsncmp( data_dir, unix_prefix, wcslen(unix_prefix) )) return FALSE;
     data_dir += 4;  /* skip \??\ prefix */
     package_dir = malloc((wcslen(data_dir) + wcslen(suffix) + 1) * sizeof(WCHAR));
     lstrcpyW( package_dir, data_dir );
@@ -864,10 +889,10 @@
     return (!skip_local && get_mono_path_local(path)) ||
         get_mono_path_registry(path) ||
         get_mono_path_datadir(path) ||
-        get_mono_path_unix(L"\\\\?\\unix" INSTALL_DATADIR "/wine/mono", path) ||
+        get_mono_path_unix(INSTALL_DATADIR "/wine/mono", path) ||
         (strcmp(INSTALL_DATADIR, "/usr/share") &&
-         get_mono_path_unix(L"\\\\?\\unix/usr/share/wine/mono", path)) ||
-        get_mono_path_unix(L"\\\\?\\unix/opt/wine/mono", path);
+         get_mono_path_unix("/usr/share/wine/mono", path)) ||
+        get_mono_path_unix("/opt/wine/mono", path);
 }
 
 struct InstalledRuntimeEnum
@@ -1720,6 +1745,44 @@
     return result;
 }
 
+static BOOL compile_assembly(const char *source, const char *target, char *target_path, DWORD target_path_len)
+{
+    static const char *csc = "C:\\windows\\Microsoft.NET\\Framework\\v2.0.50727\\csc.exe";
+    char cmdline[2 * MAX_PATH + 74], tmp[MAX_PATH], tmpdir[MAX_PATH], source_path[MAX_PATH];
+    STARTUPINFOA si = {.cb = sizeof(STARTUPINFOA)};
+    PROCESS_INFORMATION pi;
+    HANDLE file;
+    DWORD size;
+    BOOL ret;
+    LUID id;
+
+    if (!PathFileExistsA(csc)) return FALSE;
+    if (!AllocateLocallyUniqueId(&id)) return FALSE;
+
+    GetTempPathA(MAX_PATH, tmp);
+    if (!GetTempFileNameA(tmp, "assembly", id.LowPart, tmpdir)) return FALSE;
+    if (!CreateDirectoryA(tmpdir, NULL) && GetLastError() != ERROR_ALREADY_EXISTS) return FALSE;
+
+    snprintf(source_path, MAX_PATH, "%s\\source.cs", tmpdir);
+    snprintf(target_path, target_path_len, "%s\\%s", tmpdir, target);
+
+    file = CreateFileA(source_path, GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, 0);
+    if (file == INVALID_HANDLE_VALUE) return FALSE;
+    ret = WriteFile(file, source, strlen(source), &size, NULL);
+    CloseHandle(file);
+    if (!ret) return FALSE;
+
+    snprintf(cmdline, ARRAY_SIZE(cmdline), "%s /t:library /out:\"%s\" \"%s\"", csc, target_path, source_path);
+    ret = CreateProcessA(csc, cmdline, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
+    if (!ret) return FALSE;
+
+    WaitForSingleObject(pi.hProcess, INFINITE);
+    CloseHandle(pi.hThread);
+    CloseHandle(pi.hProcess);
+
+    return PathFileExistsA(target_path);
+}
+
 static MonoAssembly* CDECL mono_assembly_preload_hook_fn(MonoAssemblyName *aname, char **assemblies_path, void *user_data)
 {
     int flags = 0;
@@ -1752,6 +1815,8 @@
     static const WCHAR dotdllW[] = {'.','d','l','l',0};
     static const WCHAR dotexeW[] = {'.','e','x','e',0};
 
+    const char *sgi = getenv("SteamGameId");
+
     stringname = mono_stringify_assembly_name(aname);
     assemblyname = mono_assembly_name_get_name(aname);
     culture = mono_assembly_name_get_culture(aname);
@@ -1808,6 +1873,90 @@
         }
     }
 
+    if (!strcmp(assemblyname, "ManagedStarter"))
+    {
+        /* HACK for Mount & Blade II: Bannerlord
+         *
+         * The launcher executable uses an AssemblyResolve event handler
+         * to redirect loads of the "ManagedStarter" assembly to
+         * Bannerlord.exe. Due to Mono issue #11319, the runtime attempts
+         * to load ManagedStarter before executing the static constructor
+         * that adds this event handler. We work around this by doing the
+         * same thing in our own assembly load hook. */
+        if (sgi && !strcmp(sgi, "261550"))
+        {
+            FIXME("hack, using Bannerlord.exe\n");
+
+            result = mono_assembly_open("Bannerlord.exe", &stat);
+
+            if (result)
+                goto done;
+            else
+                ERR("Bannerlord.exe failed to load\n");
+        }
+    }
+
+    /* HACK for games which reference a type from a non-existing DLL.
+     * Native .NET framework normally gets away with it but Mono cannot
+     * due to some deeply rooted differences. */
+    if (sgi)
+    {
+        size_t i;
+
+        static const struct {
+            const char *assembly_name;
+            const char *module_name;
+            const char *appid;
+            const char *source;
+        } assembly_hacks[] = {
+            {
+                "CameraQuakeViewer",
+                "CameraQuakeViewer.dll",
+                "527280", /* Nights of Azure */
+                "namespace CQViewer { class CQMgr {} }"
+            },
+            {
+                "UnrealEdCSharp",
+                "UnrealEdCSharp.dll",
+                "317940", /* Karmaflow */
+                "namespace ContentBrowser { class IContentBrowserBackendInterface {} class Package {} } "
+            },
+            {
+                "UnrealEdCSharp",
+                "UnrealEdCSharp.dll",
+                "321360", /* Primal Carnage: Extinction */
+                "namespace ContentBrowser { class IContentBrowserBackendInterface {} class Package {} } "
+            },
+            {
+                "DockPanel",
+                "DockPanel.dll",
+                "46450", /* Grotesque Tactics: Evil Heroes  */
+                "namespace WeifenLuo.WinFormsUI { class DockPanel {} }"
+            },
+        };
+
+        for (i = 0; i < ARRAY_SIZE(assembly_hacks); ++i)
+        {
+            if (!strcmp(assemblyname, assembly_hacks[i].assembly_name) &&
+                    !strcmp(sgi, assembly_hacks[i].appid))
+            {
+                char assembly_path[MAX_PATH];
+
+                FIXME("HACK: Building %s\n", assembly_hacks[i].module_name);
+
+                if (compile_assembly(assembly_hacks[i].source, assembly_hacks[i].module_name, assembly_path, MAX_PATH))
+                    result = mono_assembly_open(assembly_path, &stat);
+                else
+                    ERR("HACK: Failed to build %s\n", assembly_hacks[i].assembly_name);
+
+                if (result)
+                    goto done;
+
+                ERR("HACK: Failed to load %s\n", assembly_hacks[i].assembly_name);
+            }
+        }
+    }
+
     if ((search_flags & ASSEMBLY_SEARCH_GAC) != 0)
     {
         stringnameW_size = MultiByteToWideChar(CP_UTF8, 0, stringname, -1, NULL, 0);
