diff --git a/dlls/dinput/joystick_hid.c b/dlls/dinput/joystick_hid.c
--- a/dlls/dinput/joystick_hid.c
+++ b/dlls/dinput/joystick_hid.c
@@ -47,6 +47,9 @@
 #include "wine/debug.h"
 #include "wine/hid.h"
 
+#define VID_LOGITECH 0x046D
+#define PID_LOGITECH_G920 0xC262
+
 WINE_DEFAULT_DEBUG_CHANNEL(dinput);
 
 DEFINE_GUID( GUID_DEVINTERFACE_WINEXINPUT,0x6c53d5fd,0x6480,0x440f,0xb6,0x18,0x47,0x67,0x50,0xc5,0xe1,0xa6 );
@@ -172,6 +175,8 @@
 struct hid_joystick
 {
     struct dinput_device base;
+    BOOL wgi_device;
+    BOOL is_gamepad;
 
     HANDLE device;
     OVERLAPPED read_ovl;
@@ -239,6 +244,51 @@
     char *set_envelope_buf;
 };
 
+struct joystick_device
+{
+    WCHAR device_path[MAX_PATH];
+};
+
+static CRITICAL_SECTION joystick_devices_crit;
+static CRITICAL_SECTION_DEBUG joystick_devices_crit_debug =
+{
+    0, 0, &joystick_devices_crit,
+    { &joystick_devices_crit_debug.ProcessLocksList, &joystick_devices_crit_debug.ProcessLocksList },
+      0, 0, { (DWORD_PTR)(__FILE__ ": joystick_devices_crit") }
+};
+static CRITICAL_SECTION joystick_devices_crit = { &joystick_devices_crit_debug, -1, 0, 0, 0, 0 };
+
+static struct joystick_device *joystick_devices;
+static unsigned int joystick_device_count;
+
+static unsigned int get_joystick_index( const WCHAR *device_path )
+{
+    unsigned int i;
+
+    EnterCriticalSection( &joystick_devices_crit );
+    for (i = 0; i < joystick_device_count; ++i)
+        if (!wcsicmp( joystick_devices[i].device_path, device_path )) break;
+
+    if (i == joystick_device_count)
+    {
+        ++joystick_device_count;
+        joystick_devices = realloc( joystick_devices, sizeof(*joystick_devices) * joystick_device_count );
+        wcscpy( joystick_devices[i].device_path, device_path );
+    }
+    LeaveCriticalSection( &joystick_devices_crit );
+    return i;
+}
+
+static BOOL get_default_joystick_device_path( WCHAR *device_path )
+{
+    BOOL ret;
+
+    EnterCriticalSection( &joystick_devices_crit );
+    if ((ret = !!joystick_device_count)) wcscpy( device_path, joystick_devices[0].device_path );
+    LeaveCriticalSection( &joystick_devices_crit );
+    return ret;
+}
+
 static inline struct hid_joystick_effect *impl_from_IDirectInputEffect( IDirectInputEffect *iface )
 {
     return CONTAINING_RECORD( iface, struct hid_joystick_effect, IDirectInputEffect_iface );
@@ -543,6 +593,8 @@
     struct hid_collection_node *node, *node_end;
     WORD version = impl->base.dinput->dwVersion;
     BOOL ret, seen_axis[6] = {0};
+    const GUID *hack_guid;
+    const WCHAR *hack_name;
     const WCHAR *tmp;
 
     button_ofs += impl->caps.NumberInputValueCaps * sizeof(LONG);
@@ -562,6 +614,8 @@
             value_ofs += (caps->usage_max - caps->usage_min + 1) * sizeof(LONG);
         else for (j = caps->usage_min; j <= caps->usage_max; ++j)
         {
+            hack_name = NULL;
+            hack_guid = NULL;
             instance.dwOfs = value_ofs;
             switch (MAKELONG(j, caps->usage_page))
             {
@@ -571,7 +625,38 @@
             case MAKELONG(HID_USAGE_GENERIC_RX, HID_USAGE_PAGE_GENERIC):
             case MAKELONG(HID_USAGE_GENERIC_RY, HID_USAGE_PAGE_GENERIC):
             case MAKELONG(HID_USAGE_GENERIC_RZ, HID_USAGE_PAGE_GENERIC):
-                set_axis_type( &instance, seen_axis, j - HID_USAGE_GENERIC_X, &axis );
+                if (!impl->wgi_device && impl->attrs.VendorID == VID_LOGITECH && impl->attrs.ProductID == PID_LOGITECH_G920)
+                {
+                    if (j == HID_USAGE_GENERIC_X)
+                    {
+                        set_axis_type( &instance, seen_axis, 0, &axis );
+                        hack_guid = &GUID_XAxis;
+                        hack_name = L"Wheel axis";
+                    }
+                    else if (j == HID_USAGE_GENERIC_Y)
+                    {
+                        set_axis_type( &instance, seen_axis, 2, &axis );
+                        hack_guid = &GUID_YAxis;
+                        hack_name = L"Accelerator";
+                    }
+                    else if (j == HID_USAGE_GENERIC_Z)
+                    {
+                        set_axis_type( &instance, seen_axis, 5, &axis );
+                        hack_guid = &GUID_RzAxis;
+                        hack_name = L"Brake";
+                    }
+                    else if (j == HID_USAGE_GENERIC_RZ)
+                    {
+                        instance.dwType = DIDFT_ABSAXIS | DIDFT_MAKEINSTANCE( 6 + axis++ );
+                        hack_guid = &GUID_Slider;
+                        hack_name = L"Clutch";
+                    }
+                    else WARN("unknown axis usage page %x usage %lx for Logitech G920\n", caps->usage_page, j);
+                }
+                else
+                {
+                    set_axis_type( &instance, seen_axis, j - HID_USAGE_GENERIC_X, &axis );
+                }
                 instance.dwFlags = DIDOI_ASPECTPOSITION;
                 break;
             case MAKELONG(HID_USAGE_SIMULATION_STEERING, HID_USAGE_PAGE_SIMULATION):
@@ -608,12 +693,16 @@
             }
             instance.wUsagePage = caps->usage_page;
             instance.wUsage = j;
-            instance.guidType = *object_usage_to_guid( instance.wUsagePage, instance.wUsage );
+            if (hack_guid)
+                instance.guidType = *hack_guid;
+            else
+                instance.guidType = *object_usage_to_guid( instance.wUsagePage, instance.wUsage );
             instance.wReportId = caps->report_id;
             instance.wCollectionNumber = caps->link_collection;
             instance.dwDimension = caps->units;
             instance.wExponent = caps->units_exp;
-            if ((tmp = object_usage_to_string( &instance ))) lstrcpynW( instance.tszName, tmp, MAX_PATH );
+            if (hack_name) lstrcpynW( instance.tszName, hack_name, MAX_PATH );
+            else if ((tmp = object_usage_to_string( &instance ))) lstrcpynW( instance.tszName, tmp, MAX_PATH );
             else swprintf( instance.tszName, MAX_PATH, L"Unknown %u", DIDFT_GETINSTANCE( instance.dwType ) );
             check_pid_effect_axis_caps( impl, &instance );
             ret = enum_object( impl, filter, flags, callback, object, caps, &instance, data );
@@ -805,13 +894,23 @@
     case (DWORD_PTR)DIPROP_JOYSTICKID:
     {
         DIPROPDWORD *value = (DIPROPDWORD *)header;
-        value->dwData = impl->base.instance.guidInstance.Data3;
+        value->dwData = get_joystick_index( impl->device_path );
         return DI_OK;
     }
     case (DWORD_PTR)DIPROP_GUIDANDPATH:
     {
         DIPROPGUIDANDPATH *value = (DIPROPGUIDANDPATH *)header;
         value->guidClass = GUID_DEVCLASS_HIDCLASS;
+
+        /* CW-Bug-Id: #23185 Emulate Steam Input native hooks for native SDL */
+        if (impl->attrs.VendorID == 0x28de && impl->attrs.ProductID == 0x11ff)
+        {
+            const WCHAR *tmp;
+            if ((tmp = wcschr( impl->device_path, '#' ))) tmp = wcschr( tmp + 1, '#' );
+            lstrcpynW( value->wszPath, impl->device_path, tmp - impl->device_path + 1 );
+            return DI_OK;
+        }
+
         lstrcpynW( value->wszPath, impl->device_path, MAX_PATH );
         return DI_OK;
     }
@@ -1125,6 +1224,7 @@
 {
     BYTE old_state[DEVICE_STATE_MAX_SIZE];
     BYTE buttons[128];
+    BOOL reset_state;
     DWORD time;
     DWORD seq;
 };
@@ -1140,6 +1240,8 @@
 
     value = params->buttons[instance->wUsage - 1];
     old_value = params->old_state[instance->dwOfs];
+    if (params->reset_state) value = 0;
+
     device->device_state[instance->dwOfs] = value;
     if (old_value != value) queue_event( iface, index, value, params->time, params->seq );
 
@@ -1224,6 +1326,23 @@
                                              instance->wUsagePage, instance->wUsage, status );
     if (instance->dwType & DIDFT_AXIS) value = scale_axis_value( logical_value, properties );
     else value = scale_value( logical_value, properties );
+
+    if (params->reset_state)
+    {
+        if (instance->dwType & DIDFT_POV) value = -1;
+        else if (instance->dwType & DIDFT_AXIS)
+        {
+            if (!properties->range_min)
+            {
+                if (impl->is_gamepad && instance->wUsagePage == HID_USAGE_PAGE_GENERIC &&
+                        (instance->wUsage == HID_USAGE_GENERIC_Z || instance->wUsage == HID_USAGE_GENERIC_RZ))
+                    value = properties->range_min;
+                else
+                    value = properties->range_max / 2;
+            }
+            else value = round( (properties->range_min + properties->range_max) / 2.0 );
+        }
+    }
 
     old_value = *(LONG *)(params->old_state + instance->dwOfs);
     *(LONG *)(impl->base.device_state + instance->dwOfs) = value;
@@ -1253,6 +1372,11 @@
     BOOL ret;
 
     ret = GetOverlappedResult( impl->device, &impl->read_ovl, &count, FALSE );
+
+    if (WaitForSingleObject(steam_overlay_event, 0) == WAIT_OBJECT_0) /* steam overlay is enabled */
+        params.reset_state = TRUE;
+    else
+        params.reset_state = FALSE;
 
     EnterCriticalSection( &impl->base.crit );
     while (ret)
@@ -1560,6 +1684,9 @@
         break;
     }
 
+    if (attrs->VendorID == VID_LOGITECH && attrs->ProductID == PID_LOGITECH_G920)
+        type = DI8DEVTYPE_DRIVING | (DI8DEVTYPEDRIVING_DUALPEDALS << 8);
+
     instance->dwDevType = device_type_for_version( type, version ) | DIDEVTYPE_HID;
     TRACE("detected device type %#lx\n", instance->dwDevType);
 
@@ -1626,6 +1753,8 @@
                                                      attrs, caps, instance, version )))
                 continue;
         }
+        /* Assign joystick index if the device path is first seen. */
+        get_joystick_index( detail->DevicePath );
 
         /* enumerate device by GUID */
         if (IsEqualGUID( guid, &instance->guidProduct ) || IsEqualGUID( guid, &instance->guidInstance )) break;
@@ -2007,6 +2136,16 @@
     return DIENUM_CONTINUE;
 }
 
+static BOOL hid_joystick_is_gamepad(DWORD type, DWORD version)
+{
+    const DWORD joystick_type = DIDEVTYPE_JOYSTICK | (DIDEVTYPEJOYSTICK_GAMEPAD << 8) | DIDEVTYPE_HID;
+
+    if (version >= 0x0800)
+        return (GET_DIDEVICE_TYPE(type) == DI8DEVTYPE_GAMEPAD);
+
+    return type == joystick_type;
+}
+
 HRESULT hid_joystick_create_device( struct dinput *dinput, const GUID *guid, IDirectInputDevice8W **out )
 {
     static const DIPROPHEADER filter =
@@ -2042,10 +2181,18 @@
     impl->base.read_event = CreateEventW( NULL, TRUE, FALSE, NULL );
 
     if (memcmp( device_path_guid.Data4, guid->Data4, sizeof(device_path_guid.Data4) ))
+    {
+        /* Let hid_joystick_device_open() populate joystick devices before checking for default joystick GUID. */
         hr = hid_joystick_device_open( -1, guid, &impl->base.instance, impl->device_path, &impl->device, &impl->preparsed,
                                        &attrs, &impl->caps, dinput->dwVersion );
+        if (hr == DIERR_DEVICENOTREG && IsEqualGUID( guid, &GUID_Joystick )
+            && get_default_joystick_device_path( impl->device_path ))
+            hr = hid_joystick_device_try_open( impl->device_path, &impl->device, &impl->preparsed, &attrs,
+                                               &impl->caps, &impl->base.instance, dinput->dwVersion );
+    }
     else
     {
+        impl->wgi_device = TRUE;
         wcscpy( impl->device_path, *(const WCHAR **)guid );
         hr = hid_joystick_device_try_open( impl->device_path, &impl->device, &impl->preparsed, &attrs,
                                            &impl->caps, &impl->base.instance, dinput->dwVersion );
@@ -2053,6 +2200,7 @@
     if (hr != DI_OK) goto failed;
 
     impl->base.caps.dwDevType = impl->base.instance.dwDevType;
+    impl->is_gamepad = hid_joystick_is_gamepad(impl->base.caps.dwDevType, dinput->dwVersion);
     impl->attrs = attrs;
     list_init( &impl->effect_list );
 
