diff --git a/dlls/win32u/imm.c b/dlls/win32u/imm.c
--- a/dlls/win32u/imm.c
+++ b/dlls/win32u/imm.c
@@ -39,7 +39,6 @@
     struct list entry;
     WORD vkey;
     WORD scan;
-    BOOL key_consumed;
     DWORD cursor_pos;
     WCHAR *comp_str;
     WCHAR *result_str;
@@ -48,6 +47,7 @@
 
 struct imc
 {
+    struct user_object obj;
     DWORD    thread_id;
     UINT_PTR client_ptr;
 };
@@ -94,7 +94,7 @@
     if (!(imc = malloc( sizeof(*imc) ))) return 0;
     imc->client_ptr = client_ptr;
     imc->thread_id = GetCurrentThreadId();
-    if (!(handle = alloc_user_handle( imc, NTUSER_OBJ_IMC )))
+    if (!(handle = alloc_user_handle( &imc->obj, NTUSER_OBJ_IMC )))
     {
         free( imc );
         return 0;
@@ -187,7 +187,7 @@
     WND *win;
     UINT ret = AICR_OK;
 
-    TRACE( "%p %p %x\n", hwnd, ctx, flags );
+    TRACE( "%p %p %x\n", hwnd, ctx, (int)flags );
 
     switch (flags)
     {
@@ -197,7 +197,7 @@
         break;
 
     default:
-        FIXME( "unknown flags 0x%x\n", flags );
+        FIXME( "unknown flags 0x%x\n", (int)flags );
         return AICR_FAILED;
     }
 
@@ -211,10 +211,11 @@
             return AICR_FAILED;
     }
 
-    if (ctx && !is_current_thread_window( hwnd )) return AICR_FAILED;
-    if (!(win = get_win_ptr( hwnd )) || win == WND_OTHER_PROCESS || win == WND_DESKTOP) return AICR_FAILED;
-
-    if (flags != IACE_IGNORENOCONTEXT || win->imc)
+    if (!(win = get_win_ptr( hwnd )) || win == WND_OTHER_PROCESS || win == WND_DESKTOP)
+        return AICR_FAILED;
+
+    if (ctx && win->tid != GetCurrentThreadId()) ret = AICR_FAILED;
+    else if (flags != IACE_IGNORENOCONTEXT || win->imc)
     {
         if (win->imc != ctx && get_focus() == hwnd) ret = AICR_FOCUS_CHANGED;
         win->imc = ctx;
@@ -274,17 +275,6 @@
     return thread_info->imm_thread_data;
 }
 
-static BOOL needs_ime_window( HWND hwnd )
-{
-    static const WCHAR imeW[] = {'I','M','E'};
-    WCHAR nameW[MAX_ATOM_LEN + 1];
-    UNICODE_STRING name = RTL_CONSTANT_STRING(nameW);
-
-    if (NtUserGetClassLongW( hwnd, GCL_STYLE ) & CS_IME) return FALSE;
-    name.Length = NtUserGetClassName( hwnd, FALSE, &name ) * sizeof(WCHAR);
-    return name.Length != sizeof(imeW) || memcmp( name.Buffer, imeW, sizeof(imeW) );
-}
-
 BOOL register_imm_window( HWND hwnd )
 {
     struct imm_thread_data *thread_data;
@@ -308,9 +298,9 @@
         UNICODE_STRING class_name = RTL_CONSTANT_STRING( imeW );
         UNICODE_STRING name = RTL_CONSTANT_STRING( default_imeW );
 
-        thread_data->default_hwnd = NtUserCreateWindowEx( 0, &class_name, NULL, &name,
+        thread_data->default_hwnd = NtUserCreateWindowEx( 0, &class_name, &class_name, &name,
                                                           WS_POPUP | WS_DISABLED | WS_CLIPSIBLINGS,
-                                                          0, 0, 1, 1, 0, 0, 0, 0, 0, 0, NULL, FALSE );
+                                                          0, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, FALSE );
     }
 
     return TRUE;
@@ -434,8 +424,9 @@
 
     *size = 0;
     user_lock();
-    while (count && (imc = next_thread_user_object( thread_id, &handle, NTUSER_OBJ_IMC )))
-    {
+    while (count && (imc = next_process_user_handle_ptr( &handle, NTUSER_OBJ_IMC )))
+    {
+        if (thread_id != -1 && imc->thread_id != thread_id) continue;
         buffer[(*size)++] = handle;
         count--;
     }
@@ -449,35 +440,16 @@
     static UINT ime_update_count;
 
     struct imm_thread_data *data = get_imm_thread_data();
-    UINT id = -1, comp_len, result_len, prev_result_len;
-    WCHAR *prev_result_str, *tmp;
+    UINT id = -1, comp_len, result_len;
     struct ime_update *update;
 
-    TRACE( "hwnd %p, cursor_pos %u - %u, comp_str %s, result_str %s\n", hwnd, LOWORD(cursor_pos),
-           HIWORD(cursor_pos), debugstr_w(comp_str), debugstr_w(result_str) );
+    TRACE( "hwnd %p, cursor_pos %u, comp_str %s, result_str %s\n", hwnd, cursor_pos,
+           debugstr_w(comp_str), debugstr_w(result_str) );
 
     comp_len = comp_str ? wcslen( comp_str ) + 1 : 0;
     result_len = result_str ? wcslen( result_str ) + 1 : 0;
 
-    /* prepend or keep the previous result string, if there was any */
-    if (!data->ime_process_vkey || !data->update) prev_result_str = NULL;
-    else prev_result_str = data->update->result_str;
-    prev_result_len = prev_result_str ? wcslen( prev_result_str ) + 1 : 0;
-
-    if (!prev_result_len && !result_len) tmp = NULL;
-    else if (!(tmp = malloc( (prev_result_len + result_len) * sizeof(WCHAR) ))) return;
-
-    if (prev_result_len && result_len) prev_result_len -= 1; /* concat both strings */
-    if (prev_result_len) memcpy( tmp, prev_result_str, prev_result_len * sizeof(WCHAR) );
-    if (result_len) memcpy( tmp + prev_result_len, result_str, result_len * sizeof(WCHAR) );
-    result_len += prev_result_len;
-    result_str = tmp;
-
-    if (!(update = malloc( offsetof(struct ime_update, buffer[comp_len + result_len]) )))
-    {
-        free( tmp );
-        return;
-    }
+    if (!(update = malloc( offsetof(struct ime_update, buffer[comp_len + result_len]) ))) return;
     update->cursor_pos = cursor_pos;
     update->comp_str = comp_str ? memcpy( update->buffer, comp_str, comp_len * sizeof(WCHAR) ) : NULL;
     update->result_str = result_str ? memcpy( update->buffer + comp_len, result_str, result_len * sizeof(WCHAR) ) : NULL;
@@ -487,7 +459,6 @@
         pthread_mutex_lock( &imm_mutex );
         id = update->scan = ++ime_update_count;
         update->vkey = VK_PROCESSKEY;
-        update->key_consumed = TRUE;
         list_add_tail( &ime_updates, &update->entry );
         pthread_mutex_unlock( &imm_mutex );
 
@@ -499,50 +470,6 @@
         free( data->update );
         data->update = update;
     }
-
-    free( tmp );
-}
-
-static UINT get_comp_clause_count( UINT comp_len, UINT cursor_begin, UINT cursor_end )
-{
-    if (cursor_begin == cursor_end || (cursor_begin == 0 && cursor_end == comp_len))
-        return 2;
-    else if (cursor_begin == 0 || cursor_end == comp_len)
-        return 3;
-    else
-        return 4;
-}
-
-static void set_comp_clause( DWORD *comp_clause, UINT comp_clause_count, UINT comp_len,
-                             UINT cursor_begin, UINT cursor_end )
-{
-    comp_clause[0] = 0;
-    switch (comp_clause_count)
-    {
-    case 2:
-        comp_clause[1] = comp_len;
-        break;
-    case 3:
-        comp_clause[1] = cursor_begin == 0 ? cursor_end : cursor_begin;
-        comp_clause[2] = comp_len;
-        break;
-    case 4:
-        comp_clause[1] = cursor_begin;
-        comp_clause[2] = cursor_end;
-        comp_clause[3] = comp_len;
-        break;
-    }
-}
-
-static void set_comp_attr( BYTE *comp_attr, UINT comp_attr_len, UINT cursor_begin, UINT cursor_end )
-{
-    if (cursor_begin == cursor_end)
-        memset( comp_attr, ATTR_INPUT, comp_attr_len );
-    else
-    {
-        memset( comp_attr, ATTR_CONVERTED, comp_attr_len );
-        memset( comp_attr + cursor_begin, ATTR_TARGET_CONVERTED, cursor_end - cursor_begin );
-    }
 }
 
 static struct ime_update *find_ime_update( WORD vkey, WORD scan )
@@ -555,11 +482,9 @@
     return NULL;
 }
 
-static UINT ime_to_tascii_ex( UINT vkey, UINT lparam, const BYTE *state, COMPOSITIONSTRING *compstr,
-                              BOOL *key_consumed, HIMC himc )
-{
-    UINT needed = sizeof(COMPOSITIONSTRING), comp_len, result_len, comp_clause_count = 0;
-    UINT cursor_begin = 0, cursor_end = 0;
+UINT ime_to_tascii_ex( UINT vkey, UINT lparam, const BYTE *state, COMPOSITIONSTRING *compstr, HIMC himc )
+{
+    UINT needed = sizeof(COMPOSITIONSTRING), comp_len, result_len;
     struct ime_update *update;
     void *dst;
 
@@ -573,24 +498,13 @@
         return STATUS_NOT_FOUND;
     }
 
-    *key_consumed = update->key_consumed;
-
     if (!update->comp_str) comp_len = 0;
     else
     {
         comp_len = wcslen( update->comp_str );
-        cursor_begin = LOWORD(update->cursor_pos);
-        cursor_end   = HIWORD(update->cursor_pos);
-
-        if (cursor_begin > comp_len) cursor_begin = comp_len;
-        if (cursor_end > comp_len) cursor_end = comp_len;
-        if (cursor_end < cursor_begin) cursor_end = cursor_begin;
-
-        comp_clause_count = get_comp_clause_count( comp_len, cursor_begin, cursor_end );
-
         needed += comp_len * sizeof(WCHAR); /* GCS_COMPSTR */
         needed += comp_len; /* GCS_COMPATTR */
-        needed += comp_clause_count * sizeof(DWORD); /* GCS_COMPCLAUSE */
+        needed += 2 * sizeof(DWORD); /* GCS_COMPCLAUSE */
     }
 
     if (!update->result_str) result_len = 0;
@@ -616,7 +530,7 @@
 
     if (update->comp_str)
     {
-        compstr->dwCursorPos = cursor_begin;
+        compstr->dwCursorPos = update->cursor_pos;
 
         compstr->dwCompStrLen = comp_len;
         compstr->dwCompStrOffset = compstr->dwSize;
@@ -624,16 +538,17 @@
         memcpy( dst, update->comp_str, compstr->dwCompStrLen * sizeof(WCHAR) );
         compstr->dwSize += compstr->dwCompStrLen * sizeof(WCHAR);
 
-        compstr->dwCompClauseLen = comp_clause_count * sizeof(DWORD);
+        compstr->dwCompClauseLen = 2 * sizeof(DWORD);
         compstr->dwCompClauseOffset = compstr->dwSize;
         dst = (BYTE *)compstr + compstr->dwCompClauseOffset;
-        set_comp_clause( dst, comp_clause_count, comp_len, cursor_begin, cursor_end );
+        *((DWORD *)dst + 0) = 0;
+        *((DWORD *)dst + 1) = compstr->dwCompStrLen;
         compstr->dwSize += compstr->dwCompClauseLen;
 
         compstr->dwCompAttrLen = compstr->dwCompStrLen;
         compstr->dwCompAttrOffset = compstr->dwSize;
         dst = (BYTE *)compstr + compstr->dwCompAttrOffset;
-        set_comp_attr( dst, compstr->dwCompAttrLen, cursor_begin, cursor_end );
+        memset( dst, ATTR_INPUT, compstr->dwCompAttrLen );
         compstr->dwSize += compstr->dwCompAttrLen;
     }
 
@@ -675,7 +590,6 @@
 
         if (data->update)
         {
-            data->update->key_consumed = res;
             pthread_mutex_lock( &imm_mutex );
             list_add_tail( &ime_updates, &data->update->entry );
             pthread_mutex_unlock( &imm_mutex );
@@ -683,11 +597,11 @@
             res = TRUE;
         }
 
-        TRACE( "processing vkey %#x, scan %#x -> %lu\n", LOWORD(wparam), HIWORD(lparam), res );
+        TRACE( "processing scan %#x, vkey %#x -> %u\n", LOWORD(wparam), HIWORD(lparam), (UINT)res );
         return res;
     }
     case WINE_IME_TO_ASCII_EX:
-        return ime_to_tascii_ex( wparam, lparam, params->state, params->compstr, params->key_consumed, params->himc );
+        return ime_to_tascii_ex( wparam, lparam, params->state, params->compstr, params->himc );
     case WINE_IME_POST_UPDATE:
         post_ime_update( hwnd, wparam, (WCHAR *)lparam, (WCHAR *)params );
         return 0;
