diff --git a/dlls/d3d8/tests/visual.c b/dlls/d3d8/tests/visual.c
--- a/dlls/d3d8/tests/visual.c
+++ b/dlls/d3d8/tests/visual.c
@@ -25,11 +25,7 @@
 
 #define COBJMACROS
 #include <d3d8.h>
-#include <winternl.h>
 #include "wine/test.h"
-
-static bool is_win64 = sizeof(void *) > sizeof(int);
-static bool new_wow64;
 
 struct vec2
 {
@@ -241,7 +237,6 @@
                     if (x < 0 || x >= 640 || y < 0 || y >= 480)
                         continue;
                     color = get_readback_color(rb, x, y);
-                    todo_wine_if(i == 0 && x_side == 0 && y_side == 1 && !is_win64 && !new_wow64)
                     ok(color == expected, "%s: Pixel (%ld, %ld) has color %08x, expected %08x.\n",
                             message, x, y, color, expected);
                 }
@@ -5135,7 +5130,7 @@
         ok(SUCCEEDED(hr), "Failed to end scene, hr %#lx.\n", hr);
 
         color = getPixelColor(device, 310, 240);
-        todo_wine_if(tests[i].vertexmode == D3DFOG_NONE) ok(color_match(color, tests[i].color_left, 1),
+        ok(color_match(color, tests[i].color_left, 1),
                 "Expected left color 0x%08x, got 0x%08x, case %u.\n", tests[i].color_left, color, i);
         color = getPixelColor(device, 330, 240);
         ok(color_match(color, tests[i].color_right, 1),
@@ -5851,77 +5846,39 @@
     color = getPixelColor(device, 320, 240);
     ok(color_match(color, 0x00ff00ff, 1), "Got unexpected color 0x%08x.\n", color);
 
-    /* Tests with managed textures. They are initially dirty.*/
-    fill_surface(surface_managed0, 0x00ffff00, D3DLOCK_READONLY);
-    fill_surface(surface_managed1, 0x00ff00ff, D3DLOCK_READONLY);
+    /* Tests with managed textures. */
+    fill_surface(surface_managed0, 0x00ff0000, 0);
+    fill_surface(surface_managed1, 0x00ff0000, 0);
     hr = IDirect3DDevice8_SetTexture(device, 0, (IDirect3DBaseTexture8 *)tex_managed);
     ok(SUCCEEDED(hr), "Failed to set texture, hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 0);
-    ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
     add_dirty_rect_test_draw(device);
     color = getPixelColor(device, 320, 240);
-    ok(color_match(color, 0x00ffff00, 1), "Got unexpected color 0x%08x.\n", color);
+    ok(color_match(color, 0x00ff0000, 1),
+            "Expected color 0x00ff0000, got 0x%08x.\n", color);
     hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
     ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
     hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 1);
     ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
     add_dirty_rect_test_draw(device);
     color = getPixelColor(device, 320, 240);
-    ok(color_match(color, 0x00ff00ff, 1), "Got unexpected color 0x%08x.\n", color);
+    ok(color_match(color, 0x00ff0000, 1), "Got unexpected color 0x%08x.\n", color);
     hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
     ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
 
-    /* Read-write map, this is uploaded. */
-    fill_surface(surface_managed0, 0x00ff0080, 0);
-    fill_surface(surface_managed1, 0x00ff8000, 0);
+    /* Managed textures also honor D3DLOCK_NO_DIRTY_UPDATE. */
+    fill_surface(surface_managed0, 0x0000ff00, D3DLOCK_NO_DIRTY_UPDATE);
+    fill_surface(surface_managed1, 0x000000ff, D3DLOCK_NO_DIRTY_UPDATE);
+    add_dirty_rect_test_draw(device);
+    color = getPixelColor(device, 320, 240);
+    ok(color_match(color, 0x00ff0000, 1),
+            "Expected color 0x00ff0000, got 0x%08x.\n", color);
+    hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
+    ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
     hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 0);
     ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
     add_dirty_rect_test_draw(device);
     color = getPixelColor(device, 320, 240);
-    ok(color_match(color, 0x00ff0080, 1), "Got unexpected color 0x%08x.\n", color);
-    hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
-    ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 1);
-    ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
-    add_dirty_rect_test_draw(device);
-    color = getPixelColor(device, 320, 240);
-    ok(color_match(color, 0x00ff8000, 1), "Got unexpected color 0x%08x.\n", color);
-    hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
-    ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
-
-    /* D3DLOCK_READONLY is honored, so this data does not get uploaded. */
-    fill_surface(surface_managed0, 0x00008000, D3DLOCK_READONLY);
-    fill_surface(surface_managed1, 0x00000080, D3DLOCK_READONLY);
-    hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 0);
-    ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
-    add_dirty_rect_test_draw(device);
-    color = getPixelColor(device, 320, 240);
-    ok(color_match(color, 0x00ff0080, 1), "Got unexpected color 0x%08x.\n", color);
-    hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
-    ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 1);
-    ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
-    add_dirty_rect_test_draw(device);
-    color = getPixelColor(device, 320, 240);
-    ok(color_match(color, 0x00ff8000, 1), "Got unexpected color 0x%08x.\n", color);
-    hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
-    ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
-
-    /* Managed textures also honor D3DLOCK_NO_DIRTY_UPDATE. */
-    fill_surface(surface_managed0, 0x0000ff00, D3DLOCK_NO_DIRTY_UPDATE);
-    fill_surface(surface_managed1, 0x000000ff, D3DLOCK_NO_DIRTY_UPDATE);
-    hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 0);
-    ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
-    add_dirty_rect_test_draw(device);
-    color = getPixelColor(device, 320, 240);
-    ok(color_match(color, 0x00ff0080, 1), "Got unexpected color 0x%08x.\n", color);
-    hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
-    ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 1);
-    ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
-    add_dirty_rect_test_draw(device);
-    color = getPixelColor(device, 320, 240);
-    ok(color_match(color, 0x00ff8000, 1), "Got unexpected color 0x%08x.\n", color);
+    ok(color_match(color, 0x00ff0000, 1), "Got unexpected color 0x%08x.\n", color);
     hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
     ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
 
@@ -5930,13 +5887,11 @@
      * tracked individually. Tested on Nvidia Kepler, other drivers untested. */
     hr = IDirect3DTexture8_AddDirtyRect(tex_managed, &part_rect);
     ok(hr == S_OK, "Failed to add dirty rect, hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetTextureStageState(device, 0, D3DTSS_MAXMIPLEVEL, 0);
-    ok(SUCCEEDED(hr), "Failed to set texture stage state, hr %#lx.\n", hr);
     add_dirty_rect_test_draw(device);
     color = getPixelColor(device, 320, 240);
     ok(color_match(color, 0x0000ff00, 1), "Got unexpected color 0x%08x.\n", color);
     color = getPixelColor(device, 1, 1);
-    ok(color_match(color, 0x00ff0080, 1), "Got unexpected color 0x%08x.\n", color);
+    ok(color_match(color, 0x00ff0000, 1), "Got unexpected color 0x%08x.\n", color);
     hr = IDirect3DDevice8_Present(device, NULL, NULL, NULL, NULL);
     ok(SUCCEEDED(hr), "Failed to present, hr %#lx.\n", hr);
 
@@ -12427,152 +12382,16 @@
     release_test_context(&context);
 }
 
-/* A test to show that lights are affected by the view matrix but not the world
- * matrix, and to make sure that lighting is invalidated when the view matrix
- * (and nothing else) changes. */
-static void test_lighting_matrices(void)
-{
-    struct d3d8_test_context context;
-    struct surface_readback rb;
-    IDirect3DDevice8 *device;
-    unsigned int color;
-    HRESULT hr;
-
-    static const D3DMATERIAL8 material = {.Power = 1.0f};
-
-    static const D3DLIGHT8 light =
-    {
-        .Type = D3DLIGHT_SPOT,
-        .Diffuse = {1.0f, 1.0f, 1.0f, 1.0f},
-        .Specular = {1.0f, 1.0f, 1.0f, 1.0f},
-        .Position = {2.0f, -2.0f, 2.0f},
-        .Direction = {1.0f, 1.0f, -0.5f},
-        .Range = 1000.0f,
-        .Falloff = 1.0f,
-        .Attenuation1 = 0.1f,
-        .Theta = 1.0f,
-        .Phi = 3.0f,
-    };
-
-    static const D3DMATRIX world_matrix =
-    {{{
-        0.5f, 0.0f, 0.0f, 0.0f,
-        0.0f, 0.5f, 0.0f, 0.0f,
-        0.0f, 0.0f, 0.5f, 0.0f,
-        0.0f, 0.2f, 0.0f, 1.0f,
-    }}};
-
-    static const D3DMATRIX view_matrix =
-    {{{
-        1.0f, 0.0f, 0.0f, 0.0f,
-        0.0f, 1.0f, 0.0f, 0.0f,
-        0.0f, 0.0f, 1.0f, 0.0f,
-        0.5f, 0.5f, 0.0f, 1.0f,
-    }}};
-
-    static const struct
-    {
-        struct vec3 position;
-        struct vec3 normal;
-        unsigned int diffuse;
-        unsigned int specular;
-    }
-    tri[] =
-    {
-        {{ 2.0f, -2.0f, 0.5f}, {-0.3f, -0.4f, 0.4f}, 0x00ff0000, 0x0000ff00},
-        {{-2.0f, -2.0f, 0.5f}, {-1.0f,  0.4f, 0.4f}, 0x00ff0000, 0x0000ff00},
-        {{ 0.0f,  2.0f, 0.5f}, { 0.0f, -0.4f, 0.4f}, 0x00ff0000, 0x0000ff00},
-    },
-    tri2[] =
-    {
-        {{ 1.0f, -3.0f, 0.5f}, {-0.3f, -0.4f, 0.4f}, 0x00ff0000, 0x0000ff00},
-        {{-3.0f, -3.0f, 0.5f}, {-1.0f,  0.4f, 0.4f}, 0x00ff0000, 0x0000ff00},
-        {{-1.0f,  1.0f, 0.5f}, { 0.0f, -0.4f, 0.4f}, 0x00ff0000, 0x0000ff00},
-    };
-
-    if (!init_test_context(&context))
-        return;
-    device = context.device;
-
-    hr = IDirect3DDevice8_SetVertexShader(device, D3DFVF_XYZ | D3DFVF_NORMAL | D3DFVF_DIFFUSE | D3DFVF_SPECULAR);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetMaterial(device, &material);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetLight(device, 123, &light);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_LightEnable(device, 123, TRUE);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetRenderState(device, D3DRS_SPECULARENABLE, TRUE);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetTransform(device, D3DTS_WORLD, &world_matrix);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    hr = IDirect3DDevice8_Clear(device, 0, NULL, D3DCLEAR_TARGET | D3DCLEAR_ZBUFFER, 0x00000000, 1.0f, 0);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_BeginScene(device);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_DrawPrimitiveUP(device, D3DPT_TRIANGLESTRIP, 1, tri, sizeof(*tri));
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_EndScene(device);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    get_surface_readback(context.backbuffer, &rb);
-    color = get_readback_color(&rb, 320, 240);
-    ok(color_match(color, 0x939100, 1), "Got color %08x.\n", color);
-    color = get_readback_color(&rb, 10, 430);
-    ok(color_match(color, 0x040400, 1), "Got color %08x.\n", color);
-    color = get_readback_color(&rb, 320, 10);
-    ok(color_match(color, 0xb5a600, 1), "Got color %08x.\n", color);
-    color = get_readback_color(&rb, 630, 430);
-    ok(color_match(color, 0xeafb00, 1), "Got color %08x.\n", color);
-    release_surface_readback(&rb);
-
-    hr = IDirect3DDevice8_Clear(device, 0, NULL, D3DCLEAR_TARGET | D3DCLEAR_ZBUFFER, 0x00000000, 1.0f, 0);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_SetTransform(device, D3DTS_VIEW, &view_matrix);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_BeginScene(device);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_DrawPrimitiveUP(device, D3DPT_TRIANGLESTRIP, 1, tri2, sizeof(*tri2));
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    hr = IDirect3DDevice8_EndScene(device);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    get_surface_readback(context.backbuffer, &rb);
-    color = get_readback_color(&rb, 320, 240);
-    ok(color_match(color, 0x1d1c00, 1), "Got color %08x.\n", color);
-    color = get_readback_color(&rb, 10, 430);
-    ok(color_match(color, 0x000000, 1), "Got color %08x.\n", color);
-    color = get_readback_color(&rb, 320, 10);
-    ok(color_match(color, 0x3f3d00, 1), "Got color %08x.\n", color);
-    color = get_readback_color(&rb, 630, 430);
-    ok(color_match(color, 0x000000, 1), "Got color %08x.\n", color);
-    release_surface_readback(&rb);
-
-    release_test_context(&context);
-}
-
 START_TEST(visual)
 {
     D3DADAPTER_IDENTIFIER8 identifier;
     IDirect3D8 *d3d;
-    BOOL is_wow64;
     HRESULT hr;
 
     if (!(d3d = Direct3DCreate8(D3D_SDK_VERSION)))
     {
         skip("Failed to create D3D8 object.\n");
         return;
-    }
-
-    if (!is_win64 && IsWow64Process( GetCurrentProcess(), &is_wow64 ) && is_wow64)
-    {
-        TEB64 *teb64 = ULongToPtr( NtCurrentTeb()->GdiBatchCount );
-        if (teb64)
-        {
-            PEB64 *peb64 = ULongToPtr( teb64->Peb );
-            new_wow64 = !!peb64->LdrData;
-        }
     }
 
     memset(&identifier, 0, sizeof(identifier));
@@ -12647,5 +12466,4 @@
     test_managed_reset();
     test_mipmap_upload();
     test_specular_shaders();
-    test_lighting_matrices();
 }
