diff --git a/dlls/ntdll/unix/security.c b/dlls/ntdll/unix/security.c
--- a/dlls/ntdll/unix/security.c
+++ b/dlls/ntdll/unix/security.c
@@ -60,7 +60,7 @@
     SID *sid;
     int primary_group = -1;
 
-    TRACE( "(%p,0x%08x,%p,%d,%p,%p,%p,%p,%p,%p,%p,%p,%p)\n", handle, access, attr,
+    TRACE( "(%p,0x%08x,%p,%d,%p,%p,%p,%p,%p,%p,%p,%p,%p)\n", handle, (int)access, attr,
             type, token_id, expire, user, groups, privs, owner, group, dacl, source );
 
     *handle = 0;
@@ -70,7 +70,7 @@
     {
         SECURITY_QUALITY_OF_SERVICE *qos = attr->SecurityQualityOfService;
         TRACE( "ObjectAttributes->SecurityQualityOfService = {%d, %d, %d, %s}\n",
-               qos->Length, qos->ImpersonationLevel, qos->ContextTrackingMode,
+               (int)qos->Length, qos->ImpersonationLevel, qos->ContextTrackingMode,
                qos->EffectiveOnly ? "TRUE" : "FALSE");
         level = qos->ImpersonationLevel;
     }
@@ -163,7 +163,7 @@
 {
     unsigned int ret;
 
-    TRACE( "(%p,0x%08x,0x%08x,%p)\n", process, access, attributes, handle );
+    TRACE( "(%p,0x%08x,0x%08x,%p)\n", process, (int)access, (int)attributes, handle );
 
     *handle = 0;
 
@@ -198,7 +198,7 @@
 {
     unsigned int ret;
 
-    TRACE( "(%p,0x%08x,%u,0x%08x,%p)\n", thread, access, self, attributes, handle );
+    TRACE( "(%p,0x%08x,%u,0x%08x,%p)\n", thread, (int)access, self, (int)attributes, handle );
 
     *handle = 0;
 
@@ -238,7 +238,7 @@
     {
         SECURITY_QUALITY_OF_SERVICE *qos = attr->SecurityQualityOfService;
         TRACE( "ObjectAttributes->SecurityQualityOfService = {%d, %d, %d, %s}\n",
-               qos->Length, qos->ImpersonationLevel, qos->ContextTrackingMode,
+               (int)qos->Length, qos->ImpersonationLevel, qos->ContextTrackingMode,
                qos->EffectiveOnly ? "TRUE" : "FALSE");
         level = qos->ImpersonationLevel;
     }
@@ -258,65 +258,6 @@
     return status;
 }
 
-static const char *debugstr_TokenInformationClass( TOKEN_INFORMATION_CLASS class )
-{
-    static const char * const names[] =
-    {
-        NULL,
-        "TokenUser",
-        "TokenGroups",
-        "TokenPrivileges",
-        "TokenOwner",
-        "TokenPrimaryGroup",
-        "TokenDefaultDacl",
-        "TokenSource",
-        "TokenType",
-        "TokenImpersonationLevel",
-        "TokenStatistics",
-        "TokenRestrictedSids",
-        "TokenSessionId",
-        "TokenGroupsAndPrivileges",
-        "TokenSessionReference",
-        "TokenSandBoxInert",
-        "TokenAuditPolicy",
-        "TokenOrigin",
-        "TokenElevationType",
-        "TokenLinkedToken",
-        "TokenElevation",
-        "TokenHasRestrictions",
-        "TokenAccessInformation",
-        "TokenVirtualizationAllowed",
-        "TokenVirtualizationEnabled",
-        "TokenIntegrityLevel",
-        "TokenUIAccess",
-        "TokenMandatoryPolicy",
-        "TokenLogonSid",
-        "TokenIsAppContainer",
-        "TokenCapabilities",
-        "TokenAppContainerSid",
-        "TokenAppContainerNumber",
-        "TokenUserClaimAttributes",
-        "TokenDeviceClaimAttributes",
-        "TokenRestrictedUserClaimAttributes",
-        "TokenRestrictedDeviceClaimAttributes",
-        "TokenDeviceGroups",
-        "TokenRestrictedDeviceGroups",
-        "TokenSecurityAttributes",
-        "TokenIsRestricted",
-        "TokenProcessTrustLevel",
-        "TokenPrivateNameSpace",
-        "TokenSingletonAttributes",
-        "TokenBnoIsolation",
-        "TokenChildProcessFlags",
-        "TokenIsLessPrivilegedAppContainer",
-        "TokenIsSandboxed",
-        "TokenIsAppSilo",
-        "TokenLoggingInformation",
-    };
-
-    if (class < ARRAY_SIZE(names) && names[class]) return names[class];
-    return wine_dbg_sprintf( "%u", class );
-}
 
 /***********************************************************************
  *             NtQueryInformationToken  (NTDLL.@)
@@ -373,7 +314,7 @@
     ULONG len = 0;
     unsigned int status = STATUS_SUCCESS;
 
-    TRACE( "(%p,%s,%p,%d,%p)\n", token, debugstr_TokenInformationClass(class), info, length, retlen );
+    TRACE( "(%p,%d,%p,%d,%p)\n", token, class, info, (int)length, retlen );
 
     if (class < MaxTokenInfoClass) len = info_len[class];
     if (retlen) *retlen = len;
@@ -382,6 +323,28 @@
     switch (class)
     {
     case TokenUser:
+        if (localsystem_sid)
+        {
+            static const struct sid local_system_sid = { SID_REVISION, 1, SECURITY_NT_AUTHORITY, { SECURITY_LOCAL_SYSTEM_RID } };
+            DWORD sid_len = offsetof( struct sid, sub_auth[local_system_sid.sub_count] );
+            TOKEN_USER *tuser;
+            PSID sid;
+
+            if (retlen) *retlen = sid_len + sizeof(TOKEN_USER);
+            if (sid_len + sizeof(TOKEN_USER) > length)
+            {
+                status = STATUS_BUFFER_TOO_SMALL;
+            }
+            else
+            {
+                tuser = info;
+                sid = tuser + 1;
+                tuser->User.Sid = sid;
+                tuser->User.Attributes = 0;
+                memcpy( sid, &local_system_sid, sid_len );
+            }
+            break;
+        }
         SERVER_START_REQ( get_token_sid )
         {
             TOKEN_USER *tuser = info;
@@ -575,7 +538,7 @@
 
             req->handle = wine_server_obj_handle( token );
             status = wine_server_call( req );
-            if (!status) *type = reply->elevation_type;
+            if (!status) *type = reply->elevation;
         }
         SERVER_END_REQ;
         break;
@@ -587,7 +550,7 @@
 
             req->handle = wine_server_obj_handle( token );
             status = wine_server_call( req );
-            if (!status) elevation->TokenIsElevated = reply->is_elevated;
+            if (!status) elevation->TokenIsElevated = (reply->elevation == TokenElevationTypeFull);
         }
         SERVER_END_REQ;
         break;
@@ -657,7 +620,7 @@
         break;
 
     default:
-        ERR( "Unhandled token information class %s\n", debugstr_TokenInformationClass(class) );
+        ERR( "Unhandled token information class %u\n", class );
         return STATUS_NOT_IMPLEMENTED;
     }
     return status;
@@ -672,7 +635,7 @@
 {
     unsigned int ret = STATUS_NOT_IMPLEMENTED;
 
-    TRACE( "%p %d %p %u\n", token, class, info, length );
+    TRACE( "%p %d %p %u\n", token, class, info, (int)length );
 
     switch (class)
     {
@@ -737,7 +700,8 @@
                                      SID_AND_ATTRIBUTES *capabilities, ULONG handle_count, HANDLE *handle )
 {
     FIXME("(%p, %p, %x, %p, %p, %u, %p, %u, %p): stub\n",
-          token_handle, token, access, attr, sid, count, capabilities, handle_count, handle );
+          token_handle, token, (int)access, attr, sid,
+          (int)count, capabilities, (int)handle_count, handle );
 
     /* we need to return a NULL handle since later it will be passed to NtClose and that must not fail */
     *token_handle = NULL;
@@ -751,7 +715,7 @@
 NTSTATUS WINAPI NtAdjustGroupsToken( HANDLE token, BOOLEAN reset, TOKEN_GROUPS *groups,
                                      ULONG length, TOKEN_GROUPS *prev, ULONG *retlen )
 {
-    FIXME( "%p %d %p %u %p %p\n", token, reset, groups, length, prev, retlen );
+    FIXME( "%p %d %p %u %p %p\n", token, reset, groups, (int)length, prev, retlen );
     return STATUS_NOT_IMPLEMENTED;
 }
 
@@ -764,7 +728,7 @@
 {
     unsigned int ret;
 
-    TRACE( "(%p,0x%08x,%p,0x%08x,%p,%p)\n", token, disable, privs, length, prev, retlen );
+    TRACE( "(%p,0x%08x,%p,0x%08x,%p,%p)\n", token, disable, privs, (int)length, prev, retlen );
 
     SERVER_START_REQ( adjust_token_privileges )
     {
@@ -800,11 +764,11 @@
     SID *sids = NULL;
     unsigned int status;
 
-    TRACE( "%p %#x %p %p %p %p\n", token, flags, disable_sids, privileges,
+    TRACE( "%p %#x %p %p %p %p\n", token, (int)flags, disable_sids, privileges,
            restrict_sids, new_token );
 
     if (flags)
-        FIXME( "flags %#x unsupported\n", flags );
+        FIXME( "flags %#x unsupported\n", (int)flags );
 
     if (restrict_sids)
         FIXME( "support for restricting sids not yet implemented\n" );
@@ -897,7 +861,7 @@
     ULONG priv_len;
 
     TRACE( "(%p, %p, %08x, %p, %p, %p, %p, %p)\n",
-           descr, token, access, mapping, privs, retlen, access_granted, access_status );
+           descr, token, (int)access, mapping, privs, retlen, access_granted, access_status );
 
     if (!privs || !retlen) return STATUS_ACCESS_VIOLATION;
     priv_len = *retlen;
@@ -944,42 +908,12 @@
                                             UNICODE_STRING *typename, UNICODE_STRING *objectname,
                                             PSECURITY_DESCRIPTOR descr, ACCESS_MASK access,
                                             GENERIC_MAPPING *mapping, BOOLEAN creation,
-                                            ACCESS_MASK *access_granted, NTSTATUS *access_status,
+                                            ACCESS_MASK *access_granted, BOOLEAN *access_status,
                                             BOOLEAN *onclose )
 {
     FIXME( "(%s, %p, %s, %p, 0x%08x, %p, %d, %p, %p, %p), stub\n",
-           debugstr_us(subsystem), handle, debugstr_us(typename), descr, access,
+           debugstr_us(subsystem), handle, debugstr_us(typename), descr, (int)access,
            mapping, creation, access_granted, access_status, onclose );
-    return STATUS_NOT_IMPLEMENTED;
-}
-
-
-/***********************************************************************
- *             NtAccessCheckByTypeAndAuditAlarm  (NTDLL.@)
- */
-NTSTATUS WINAPI NtAccessCheckByTypeAndAuditAlarm( UNICODE_STRING *subsystem, HANDLE handle,
-                                                  UNICODE_STRING *typename, UNICODE_STRING *objectname,
-                                                  PSECURITY_DESCRIPTOR descr, PSID sid, ACCESS_MASK access,
-                                                  AUDIT_EVENT_TYPE audit_type, ULONG flags,
-                                                  OBJECT_TYPE_LIST *obj_list, ULONG list_len,
-                                                  GENERIC_MAPPING *mapping, BOOLEAN creation,
-                                                  ACCESS_MASK *access_granted, NTSTATUS *access_status,
-                                                  BOOLEAN *onclose )
-{
-    FIXME( "(%s, %p, %s, %s, %p, 0x%08x, %u, %x, %p, %u, %p, %d, %p, %p, %p), stub\n",
-           debugstr_us(subsystem), handle, debugstr_us(typename), debugstr_us(objectname),
-           descr, access, audit_type, flags, obj_list, list_len,
-           mapping, creation, access_granted, access_status, onclose );
-    return STATUS_NOT_IMPLEMENTED;
-}
-
-
-/***********************************************************************
- *             NtCloseObjectAuditAlarm  (NTDLL.@)
- */
-NTSTATUS WINAPI NtCloseObjectAuditAlarm( UNICODE_STRING *subsystem, HANDLE handle, BOOLEAN onclose )
-{
-    FIXME( "%s %p %u: stub\n", debugstr_us(subsystem), handle, onclose );
     return STATUS_NOT_IMPLEMENTED;
 }
 
@@ -995,7 +929,7 @@
     void *buffer;
     unsigned int buffer_size = 512;
 
-    TRACE( "(%p,0x%08x,%p,0x%08x,%p)\n", handle, info, descr, length, retlen );
+    TRACE( "(%p,0x%08x,%p,0x%08x,%p)\n", handle, (int)info, descr, (int)length, retlen );
 
     for (;;)
     {
@@ -1055,7 +989,7 @@
     OBJECT_ATTRIBUTES attr;
     unsigned int status;
 
-    TRACE( "%p 0x%08x %p\n", handle, info, descr );
+    TRACE( "%p 0x%08x %p\n", handle, (int)info, descr );
 
     if (!descr) return STATUS_ACCESS_VIOLATION;
 
