diff --git a/dlls/msvcrt/math.c b/dlls/msvcrt/math.c
--- a/dlls/msvcrt/math.c
+++ b/dlls/msvcrt/math.c
@@ -40,7 +40,6 @@
 #include <stdio.h>
 #include <fenv.h>
 #include <fpieee.h>
-#include <inttypes.h>
 #include <limits.h>
 #include <locale.h>
 #include <math.h>
@@ -73,6 +72,16 @@
     sse2_supported = IsProcessorFeaturePresent( PF_XMMI64_INSTRUCTIONS_AVAILABLE );
 #if _MSVCR_VER <=71
     sse2_enabled = FALSE;
+    {
+        char sgi[64];
+
+        if (GetEnvironmentVariableA("SteamGameId", sgi, sizeof(sgi))
+                && (!strcmp(sgi, "560430") || !strcmp(sgi, "12330")))
+        {
+            sse2_supported = FALSE;
+            FIXME("HACK: disabling sse2 support in msvcrt.\n");
+        }
+    }
 #else
     sse2_enabled = sse2_supported;
 #endif
@@ -469,7 +478,7 @@
 }
 #endif
 
-#ifdef __i386__
+#if (defined(__GNUC__) || defined(__clang__)) && defined(__i386__)
 
 #define CREATE_FPU_FUNC1(name, call) \
     __ASM_GLOBAL_FUNC(name, \
@@ -583,7 +592,7 @@
         __ASM_CFI(".cfi_same_value %ebp\n\t")
         "ret")
 
-#endif /* __i386__ */
+#endif /* (defined(__GNUC__) || defined(__clang__)) && defined(__i386__) */
 
 /*********************************************************************
  *		_fpclass (MSVCRT.@)
@@ -732,6 +741,7 @@
 static void _setfp_sse( unsigned int *cw, unsigned int cw_mask,
         unsigned int *sw, unsigned int sw_mask )
 {
+#if defined(__GNUC__) || defined(__clang__)
     unsigned int old_fpword, fpword = get_mxcsr();
     unsigned int flags;
 
@@ -815,13 +825,18 @@
     }
 
     if (fpword != old_fpword) set_mxcsr( fpword );
+#else
+    FIXME("not implemented\n");
+    if (cw) *cw = 0;
+    if (sw) *sw = 0;
+#endif
 }
 #endif
 
 static void _setfp( unsigned int *cw, unsigned int cw_mask,
         unsigned int *sw, unsigned int sw_mask )
 {
-#ifdef __i386__
+#if (defined(__GNUC__) || defined(__clang__)) && defined(__i386__)
     unsigned long oldcw = 0, newcw = 0;
     unsigned long oldsw = 0, newsw = 0;
     unsigned int flags;
@@ -1409,6 +1424,43 @@
 }
 
 /*********************************************************************
+ *		feupdateenv (MSVCR120.@)
+ */
+int CDECL feupdateenv(const fenv_t *env)
+{
+    fenv_t set;
+    fegetenv(&set);
+    set._Fe_ctl = env->_Fe_ctl;
+    set._Fe_stat |= env->_Fe_stat;
+    return fesetenv(&set);
+}
+
+/*********************************************************************
+ *      fetestexcept (MSVCR120.@)
+ */
+int CDECL fetestexcept(int flags)
+{
+    return _statusfp() & flags;
+}
+
+/*********************************************************************
+ *      fesetexceptflag (MSVCR120.@)
+ */
+int CDECL fesetexceptflag(const fexcept_t *status, int excepts)
+{
+    fenv_t env;
+
+    excepts &= FE_ALL_EXCEPT;
+    if(!excepts)
+        return 0;
+
+    fegetenv(&env);
+    env._Fe_stat &= ~fenv_encode(excepts, excepts);
+    env._Fe_stat |= *status & fenv_encode(excepts, excepts);
+    return fesetenv(&env);
+}
+
+/*********************************************************************
  *      feraiseexcept (MSVCR120.@)
  */
 int CDECL feraiseexcept(int flags)
@@ -1418,40 +1470,6 @@
     flags &= FE_ALL_EXCEPT;
     fegetenv(&env);
     env._Fe_stat |= fenv_encode(flags, flags);
-    return fesetenv(&env);
-}
-
-/*********************************************************************
- *		feupdateenv (MSVCR120.@)
- */
-int CDECL feupdateenv(const fenv_t *env)
-{
-    int except = fetestexcept(FE_ALL_EXCEPT);
-    return fesetenv(env) || feraiseexcept(except);
-}
-
-/*********************************************************************
- *      fetestexcept (MSVCR120.@)
- */
-int CDECL fetestexcept(int flags)
-{
-    return _statusfp() & flags;
-}
-
-/*********************************************************************
- *      fesetexceptflag (MSVCR120.@)
- */
-int CDECL fesetexceptflag(const fexcept_t *status, int excepts)
-{
-    fenv_t env;
-
-    excepts &= FE_ALL_EXCEPT;
-    if(!excepts)
-        return 0;
-
-    fegetenv(&env);
-    env._Fe_stat &= ~fenv_encode(excepts, excepts);
-    env._Fe_stat |= *status & fenv_encode(excepts, excepts);
     return fesetenv(&env);
 }
 
@@ -1540,7 +1558,7 @@
  */
 void CDECL _fpreset(void)
 {
-#ifdef __i386__
+#if (defined(__GNUC__) || defined(__clang__)) && defined(__i386__)
     const unsigned int x86_cw = 0x27f;
     __asm__ __volatile__( "fninit; fldcw %0" : : "m" (x86_cw) );
     if (sse2_supported)
@@ -1616,6 +1634,23 @@
 #if _MSVCR_VER>=120
 
 /*********************************************************************
+ *      rint (MSVCR120.@)
+ */
+double CDECL MSVCRT_rint(double x)
+{
+    unsigned cw;
+    double y;
+
+    cw = _controlfp(0, 0);
+    if ((cw & _MCW_PC) != _PC_53)
+        _controlfp(_PC_53, _MCW_PC);
+    y = rint(x);
+    if ((cw & _MCW_PC) != _PC_53)
+        _controlfp(cw, _MCW_PC);
+    return y;
+}
+
+/*********************************************************************
  *		_nearbyint (MSVCR120.@)
  *
  * Based on musl: src/math/nearbyteint.c
@@ -1633,7 +1668,7 @@
         cw |= _EM_INEXACT;
         _setfp(&cw, _EM_INEXACT, NULL, 0);
     }
-    x = rint(x);
+    x = MSVCRT_rint(x);
     if (update_cw || update_sw)
     {
         sw = 0;
@@ -2069,20 +2104,6 @@
 }
 #endif
 
-#if _MSVCR_VER>=120
-/*********************************************************************
- *              imaxdiv (MSVCR100.@)
- */
-imaxdiv_t CDECL imaxdiv(intmax_t num, intmax_t denom)
-{
-  imaxdiv_t ret;
-
-  ret.quot = num / denom;
-  ret.rem = num % denom;
-  return ret;
-}
-#endif
-
 #ifdef __i386__
 
 /*********************************************************************
@@ -2580,7 +2601,7 @@
 {
     double d;
 
-    d = rint(x);
+    d = MSVCRT_rint(x);
     if ((d < 0 && d != (double)(__msvcrt_long)d)
             || (d >= 0 && d != (double)(__msvcrt_ulong)d)) {
         *_errno() = EDOM;
@@ -2612,7 +2633,7 @@
 {
     double d;
 
-    d = rint(x);
+    d = MSVCRT_rint(x);
     if ((d < 0 && d != (double)(__int64)d)
             || (d >= 0 && d != (double)(unsigned __int64)d)) {
         *_errno() = EDOM;
@@ -2947,58 +2968,16 @@
     return res;
 }
 
-_Dcomplex CDECL _Cbuild(double r, double i)
-{
-    _Dcomplex ret;
-    ret._Val[0] = r;
-    ret._Val[1] = i;
+_Dcomplex* CDECL _Cbuild(_Dcomplex *ret, double r, double i)
+{
+    ret->_Val[0] = r;
+    ret->_Val[1] = i;
     return ret;
 }
 
-double CDECL creal(_Dcomplex z)
+double CDECL MSVCR120_creal(_Dcomplex z)
 {
     return z._Val[0];
 }
 
-double CDECL cimag(_Dcomplex z)
-{
-    return z._Val[1];
-}
-
-#if !defined(__i386__) || defined(__MINGW32__) || defined(_MSC_VER)
-_Fcomplex CDECL _FCbuild(float r, float i)
-{
-    _Fcomplex ret;
-    ret._Val[0] = r;
-    ret._Val[1] = i;
-    return ret;
-}
-#else
-#undef _FCbuild
-ULONGLONG CDECL _FCbuild(float r, float i)
-{
-    union
-    {
-        _Fcomplex c;
-        ULONGLONG ull;
-    } ret;
-
-    C_ASSERT(sizeof(_Fcomplex) == sizeof(ULONGLONG));
-
-    ret.c._Val[0] = r;
-    ret.c._Val[1] = i;
-    return ret.ull;
-}
-#endif
-
-float CDECL crealf(_Fcomplex z)
-{
-    return z._Val[0];
-}
-
-float CDECL cimagf(_Fcomplex z)
-{
-    return z._Val[1];
-}
-
 #endif /* _MSVCR_VER>=120 */
