diff --git a/dlls/shlwapi/ordinal.c b/dlls/shlwapi/ordinal.c
--- a/dlls/shlwapi/ordinal.c
+++ b/dlls/shlwapi/ordinal.c
@@ -289,7 +289,9 @@
  */
 HRESULT WINAPI RegisterDefaultAcceptHeaders(LPBC lpBC, IUnknown *lpUnknown)
 {
-  static const WCHAR szProperty[] = L"{D0FCA420-D3F5-11CF-B211-00AA004AE837}";
+  static const WCHAR szProperty[] = { '{','D','0','F','C','A','4','2','0',
+      '-','D','3','F','5','-','1','1','C','F', '-','B','2','1','1','-','0',
+      '0','A','A','0','0','4','A','E','8','3','7','}','\0' };
   BSTR property;
   IEnumFORMATETC* pIEnumFormatEtc = NULL;
   VARIANTARG var;
@@ -489,11 +491,13 @@
 {
   WCHAR xguid[40];
   INT iLen;
+  static const WCHAR wszFormat[] = {'{','%','0','8','l','X','-','%','0','4','X','-','%','0','4','X','-',
+      '%','0','2','X','%','0','2','X','-','%','0','2','X','%','0','2','X','%','0','2','X','%','0','2',
+      'X','%','0','2','X','%','0','2','X','}',0};
 
   TRACE("(%s,%p,%d)\n", debugstr_guid(guid), lpszDest, cchMax);
 
-  swprintf(xguid, ARRAY_SIZE(xguid), L"{%08lX-%04X-%04X-%02X%02X-%02X%02X%02X%02X%02X%02X}",
-          guid->Data1, guid->Data2, guid->Data3,
+  swprintf(xguid, ARRAY_SIZE(xguid), wszFormat, guid->Data1, guid->Data2, guid->Data3,
           guid->Data4[0], guid->Data4[1], guid->Data4[2], guid->Data4[3],
           guid->Data4[4], guid->Data4[5], guid->Data4[6], guid->Data4[7]);
 
@@ -548,10 +552,31 @@
  */
 BOOL WINAPI SHAboutInfoW(LPWSTR lpszDest, DWORD dwDestLen)
 {
-  static const WCHAR szIEKey[] = L"SOFTWARE\\Microsoft\\Internet Explorer";
-  static const WCHAR szWinNtKey[] = L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion";
-  static const WCHAR szWinKey[] = L"SOFTWARE\\Microsoft\\Windows\\CurrentVersion";
-  static const WCHAR szRegKey[] = L"SOFTWARE\\Microsoft\\Internet Explorer\\Registration";
+  static const WCHAR szIEKey[] = { 'S','O','F','T','W','A','R','E','\\',
+    'M','i','c','r','o','s','o','f','t','\\','I','n','t','e','r','n','e','t',
+    ' ','E','x','p','l','o','r','e','r','\0' };
+  static const WCHAR szWinNtKey[] = { 'S','O','F','T','W','A','R','E','\\',
+    'M','i','c','r','o','s','o','f','t','\\','W','i','n','d','o','w','s',' ',
+    'N','T','\\','C','u','r','r','e','n','t','V','e','r','s','i','o','n','\0' };
+  static const WCHAR szWinKey[] = { 'S','O','F','T','W','A','R','E','\\',
+    'M','i','c','r','o','s','o','f','t','\\','W','i','n','d','o','w','s','\\',
+    'C','u','r','r','e','n','t','V','e','r','s','i','o','n','\0' };
+  static const WCHAR szRegKey[] = { 'S','O','F','T','W','A','R','E','\\',
+    'M','i','c','r','o','s','o','f','t','\\','I','n','t','e','r','n','e','t',
+    ' ','E','x','p','l','o','r','e','r','\\',
+    'R','e','g','i','s','t','r','a','t','i','o','n','\0' };
+  static const WCHAR szVersion[] = { 'V','e','r','s','i','o','n','\0' };
+  static const WCHAR szCustomized[] = { 'C','u','s','t','o','m','i','z','e','d',
+    'V','e','r','s','i','o','n','\0' };
+  static const WCHAR szOwner[] = { 'R','e','g','i','s','t','e','r','e','d',
+    'O','w','n','e','r','\0' };
+  static const WCHAR szOrg[] = { 'R','e','g','i','s','t','e','r','e','d',
+    'O','r','g','a','n','i','z','a','t','i','o','n','\0' };
+  static const WCHAR szProduct[] = { 'P','r','o','d','u','c','t','I','d','\0' };
+  static const WCHAR szUpdate[] = { 'I','E','A','K',
+    'U','p','d','a','t','e','U','r','l','\0' };
+  static const WCHAR szHelp[] = { 'I','E','A','K',
+    'H','e','l','p','S','t','r','i','n','g','\0' };
   WCHAR buff[2084];
   HKEY hReg;
   DWORD dwType, dwLen;
@@ -571,25 +596,25 @@
   /* OS Version */
   buff[0] = '\0';
   dwLen = 30;
-  if (!SHGetValueW(HKEY_LOCAL_MACHINE, szIEKey, L"Version", &dwType, buff, &dwLen))
+  if (!SHGetValueW(HKEY_LOCAL_MACHINE, szIEKey, szVersion, &dwType, buff, &dwLen))
   {
     DWORD dwStrLen = lstrlenW(buff);
     dwLen = 30 - dwStrLen;
     SHGetValueW(HKEY_LOCAL_MACHINE, szIEKey,
-                L"CustomizedVersion", &dwType, buff+dwStrLen, &dwLen);
+                szCustomized, &dwType, buff+dwStrLen, &dwLen);
   }
   StrCatBuffW(lpszDest, buff, dwDestLen);
 
   /* ~Registered Owner */
   buff[0] = '~';
   dwLen = 256;
-  if (SHGetValueW(hReg, L"RegisteredOwner", 0, &dwType, buff+1, &dwLen))
+  if (SHGetValueW(hReg, szOwner, 0, &dwType, buff+1, &dwLen))
     buff[1] = '\0';
   StrCatBuffW(lpszDest, buff, dwDestLen);
 
   /* ~Registered Organization */
   dwLen = 256;
-  if (SHGetValueW(hReg, L"RegisteredOrganization", 0, &dwType, buff+1, &dwLen))
+  if (SHGetValueW(hReg, szOrg, 0, &dwType, buff+1, &dwLen))
     buff[1] = '\0';
   StrCatBuffW(lpszDest, buff, dwDestLen);
 
@@ -601,19 +626,19 @@
 
   /* ~Product Id */
   dwLen = 256;
-  if (SHGetValueW(HKEY_LOCAL_MACHINE, szRegKey, L"ProductId", &dwType, buff+1, &dwLen))
+  if (SHGetValueW(HKEY_LOCAL_MACHINE, szRegKey, szProduct, &dwType, buff+1, &dwLen))
     buff[1] = '\0';
   StrCatBuffW(lpszDest, buff, dwDestLen);
 
   /* ~IE Update Url */
   dwLen = 2048;
-  if(SHGetValueW(HKEY_LOCAL_MACHINE, szWinKey, L"IEAKUpdateUrl", &dwType, buff+1, &dwLen))
+  if(SHGetValueW(HKEY_LOCAL_MACHINE, szWinKey, szUpdate, &dwType, buff+1, &dwLen))
     buff[1] = '\0';
   StrCatBuffW(lpszDest, buff, dwDestLen);
 
   /* ~IE Help String */
   dwLen = 256;
-  if(SHGetValueW(hReg, L"IEAKHelpString", 0, &dwType, buff+1, &dwLen))
+  if(SHGetValueW(hReg, szHelp, 0, &dwType, buff+1, &dwLen))
     buff[1] = '\0';
   StrCatBuffW(lpszDest, buff, dwDestLen);
 
@@ -2092,6 +2117,12 @@
 
 #define SHELL_NO_POLICY 0xffffffff
 
+/* default shell policy registry key */
+static const WCHAR strRegistryPolicyW[] = {'S','o','f','t','w','a','r','e','\\','M','i','c','r','o',
+                                      's','o','f','t','\\','W','i','n','d','o','w','s','\\',
+                                      'C','u','r','r','e','n','t','V','e','r','s','i','o','n',
+                                      '\\','P','o','l','i','c','i','e','s',0};
+
 /*************************************************************************
  * @                          [SHLWAPI.271]
  *
@@ -2111,7 +2142,7 @@
 	HKEY hKey;
 
 	if (!lpSubKey)
-	  lpSubKey = L"Software\\Microsoft\\Windows\\CurrentVersion\\Policies";
+	  lpSubKey = strRegistryPolicyW;
 
 	retval = RegOpenKeyW(HKEY_LOCAL_MACHINE, lpSubKey, &hKey);
         if (retval != ERROR_SUCCESS)
@@ -3629,7 +3660,8 @@
  */
 BOOL WINAPI SHSkipJunction(IBindCtx *pbc, const CLSID *pclsid)
 {
-  static WCHAR szSkipBinding[] = L"Skip Binding CLSID";
+  static WCHAR szSkipBinding[] = { 'S','k','i','p',' ',
+    'B','i','n','d','i','n','g',' ','C','L','S','I','D','\0' };
   BOOL bRet = FALSE;
 
   if (pbc)
@@ -3983,10 +4015,18 @@
 }
 
 /***********************************************************************
- *              ShellMessageBoxW [SHLWAPI.388]
- */
-INT WINAPIV ShellMessageBoxW( HINSTANCE hInstance, HWND hWnd, LPCWSTR lpText,
-	                      LPCWSTR lpCaption, UINT uType, ...)
+ *              ShellMessageBoxWrapW [SHLWAPI.388]
+ *
+ * See shell32.ShellMessageBoxW
+ *
+ * NOTE:
+ * shlwapi.ShellMessageBoxWrapW is a duplicate of shell32.ShellMessageBoxW
+ * because we can't forward to it in the .spec file since it's exported by
+ * ordinal. If you change the implementation here please update the code in
+ * shell32 as well.
+ */
+INT WINAPIV ShellMessageBoxWrapW(HINSTANCE hInstance, HWND hWnd, LPCWSTR lpText,
+                                 LPCWSTR lpCaption, UINT uType, ...)
 {
     WCHAR *szText = NULL, szTitle[100];
     LPCWSTR pszText, pszTitle = szTitle;
@@ -4031,43 +4071,6 @@
     ret = MessageBoxW(hWnd, pszTemp, pszTitle, uType);
 
     free(szText);
-    LocalFree(pszTemp);
-    return ret;
-}
-
-/*************************************************************************
- *              ShellMessageBoxA [SHLWAPI.@]
- */
-int WINAPIV ShellMessageBoxA( HINSTANCE hInstance, HWND hWnd, LPCSTR lpText,
-	                      LPCSTR lpCaption, UINT uType, ...)
-{
-    char szText[100],szTitle[100];
-    LPCSTR pszText = szText, pszTitle = szTitle;
-    LPSTR pszTemp;
-    va_list args;
-    int	ret;
-
-    va_start(args, uType);
-
-    TRACE("(%p,%p,%p,%p,%08x)\n",
-	   hInstance,hWnd,lpText,lpCaption,uType);
-
-    if (IS_INTRESOURCE(lpCaption))
-        LoadStringA(hInstance, LOWORD(lpCaption), szTitle, sizeof(szTitle));
-    else
-        pszTitle = lpCaption;
-
-    if (IS_INTRESOURCE(lpText))
-        LoadStringA(hInstance, LOWORD(lpText), szText, sizeof(szText));
-    else
-        pszText = lpText;
-
-    FormatMessageA(FORMAT_MESSAGE_ALLOCATE_BUFFER | FORMAT_MESSAGE_FROM_STRING,
-                    pszText, 0, 0, (LPSTR)&pszTemp, 0, &args);
-
-    va_end(args);
-
-    ret = MessageBoxA(hWnd,pszTemp,pszTitle,uType);
     LocalFree(pszTemp);
     return ret;
 }
@@ -4301,6 +4304,9 @@
     /* first of all date */
     if (fmt_flags & (FDTF_LONGDATE | FDTF_SHORTDATE))
     {
+        static const WCHAR sep1[] = {',',' ',0};
+        static const WCHAR sep2[] = {' ',0};
+
         DWORD date = fmt_flags & FDTF_LONGDATE ? DATE_LONGDATE : DATE_SHORTDATE;
         ret = GetDateFormatW(LOCALE_USER_DEFAULT, date, &st, NULL, buf, size);
         if (ret >= size) return ret;
@@ -4310,12 +4316,12 @@
         {
             if ((fmt_flags & FDTF_LONGDATE) && (ret < size + 2))
             {
-                lstrcatW(&buf[ret-1], L", ");
+                lstrcatW(&buf[ret-1], sep1);
                 ret += 2;
             }
             else
             {
-                lstrcatW(&buf[ret-1], L" ");
+                lstrcatW(&buf[ret-1], sep2);
                 ret++;
             }
         }
@@ -4481,23 +4487,34 @@
 };
 
 /* expected to be sorted by name */
-#define OBJCOMPAT_ENTRY(name) { L""#name, OBJCOMPAT_##name }
 static const struct objcompat_entry objcompat_table[] = {
-    OBJCOMPAT_ENTRY(COCREATESHELLFOLDERONLY),
-    OBJCOMPAT_ENTRY(CTXMENU_LIMITEDQI),
-    OBJCOMPAT_ENTRY(CTXMENU_NOVERBS),
-    OBJCOMPAT_ENTRY(CTXMENU_XPQCMFLAGS),
-    OBJCOMPAT_ENTRY(NEEDSFILESYSANCESTOR),
-    OBJCOMPAT_ENTRY(NEEDSSTORAGEANCESTOR),
-    OBJCOMPAT_ENTRY(NOIPROPERTYSTORE),
-    OBJCOMPAT_ENTRY(NOLEGACYWEBVIEW),
-    OBJCOMPAT_ENTRY(NOTAFILESYSTEM),
-    OBJCOMPAT_ENTRY(NO_WEBVIEW),
-    OBJCOMPAT_ENTRY(OTNEEDSSFCACHE),
-    OBJCOMPAT_ENTRY(PINDLL),
-    OBJCOMPAT_ENTRY(UNBINDABLE),
+    { {'C','O','C','R','E','A','T','E','S','H','E','L','L','F','O','L','D','E','R','O','N','L','Y',0},
+      OBJCOMPAT_COCREATESHELLFOLDERONLY },
+    { {'C','T','X','M','E','N','U','_','L','I','M','I','T','E','D','Q','I',0},
+      OBJCOMPAT_CTXMENU_LIMITEDQI },
+    { {'C','T','X','M','E','N','U','_','N','O','V','E','R','B','S',0},
+      OBJCOMPAT_CTXMENU_LIMITEDQI },
+    { {'C','T','X','M','E','N','U','_','X','P','Q','C','M','F','L','A','G','S',0},
+      OBJCOMPAT_CTXMENU_XPQCMFLAGS },
+    { {'N','E','E','D','S','F','I','L','E','S','Y','S','A','N','C','E','S','T','O','R',0},
+      OBJCOMPAT_NEEDSFILESYSANCESTOR },
+    { {'N','E','E','D','S','S','T','O','R','A','G','E','A','N','C','E','S','T','O','R',0},
+      OBJCOMPAT_NEEDSSTORAGEANCESTOR },
+    { {'N','O','I','P','R','O','P','E','R','T','Y','S','T','O','R','E',0},
+      OBJCOMPAT_NOIPROPERTYSTORE },
+    { {'N','O','L','E','G','A','C','Y','W','E','B','V','I','E','W',0},
+      OBJCOMPAT_NOLEGACYWEBVIEW },
+    { {'N','O','T','A','F','I','L','E','S','Y','S','T','E','M',0},
+      OBJCOMPAT_NOTAFILESYSTEM },
+    { {'N','O','_','W','E','B','V','I','E','W',0},
+      OBJCOMPAT_NO_WEBVIEW },
+    { {'O','T','N','E','E','D','S','S','F','C','A','C','H','E',0},
+      OBJCOMPAT_OTNEEDSSFCACHE },
+    { {'P','I','N','D','L','L',0},
+      OBJCOMPAT_PINDLL },
+    { {'U','N','B','I','N','D','A','B','L','E',0},
+      OBJCOMPAT_UNBINDABLE }
 };
-#undef OBJCOMPAT_ENTRY
 
 /**************************************************************************
  *  SHGetObjectCompatFlags (SHLWAPI.476)
@@ -4515,7 +4532,10 @@
 DWORD WINAPI SHGetObjectCompatFlags(IUnknown *pUnk, const CLSID *clsid)
 {
     static const WCHAR compatpathW[] =
-        L"Software\\Microsoft\\Windows\\CurrentVersion\\ShellCompatibility\\Objects\\%s";
+        {'S','o','f','t','w','a','r','e','\\','M','i','c','r','o','s','o','f','t','\\',
+         'W','i','n','d','o','w','s','\\','C','u','r','r','e','n','t','V','e','r','s','i','o','n','\\',
+         'S','h','e','l','l','C','o','m','p','a','t','i','b','i','l','i','t','y','\\',
+         'O','b','j','e','c','t','s','\\','%','s',0};
     WCHAR strW[ARRAY_SIZE(compatpathW) + 38 /* { CLSID } */];
     DWORD ret, length = ARRAY_SIZE(strW);
     OLECHAR *clsid_str;
