diff --git a/dlls/iphlpapi/tests/iphlpapi.c b/dlls/iphlpapi/tests/iphlpapi.c
--- a/dlls/iphlpapi/tests/iphlpapi.c
+++ b/dlls/iphlpapi/tests/iphlpapi.c
@@ -2211,6 +2211,7 @@
    we make, and associates it with our process. */
 static void test_GetExtendedTcpTable_owner( int family )
 {
+    BOOL found_it;
     SOCKET sock;
     int port;
     DWORD i, ret;
@@ -2266,7 +2267,7 @@
     if (family == AF_INET)
     {
         MIB_TCPTABLE_OWNER_PID *table = raw_table;
-        BOOL found_it = FALSE;
+        found_it = FALSE;
         for (i = 0; i < table->dwNumEntries; i++)
         {
             MIB_TCPROW_OWNER_PID *row = &table->table[i];
@@ -2283,7 +2284,7 @@
     else
     {
         MIB_TCP6TABLE_OWNER_PID *table = raw_table;
-        BOOL found_it = FALSE;
+        found_it = FALSE;
         for (i = 0; i < table->dwNumEntries; i++)
         {
             MIB_TCP6ROW_OWNER_PID *row = &table->table[i];
@@ -2293,6 +2294,186 @@
                 ok( row->dwOwningPid == GetCurrentProcessId(), "unexpected socket owner %04lx\n", row->dwOwningPid );
                 found_it = TRUE;
                 break;
+            }
+        }
+        ok( found_it, "no table entry for socket\n" );
+    }
+
+    free( raw_table );
+    raw_table = NULL;
+    ret = get_extended_tcp_table( family, TCP_TABLE_OWNER_MODULE_ALL, &raw_table );
+    if (ret != ERROR_SUCCESS)
+    {
+        skip( "error %lu getting TCP table\n", ret );
+        goto done;
+    }
+
+    if (family == AF_INET)
+    {
+        MIB_TCPTABLE_OWNER_MODULE *t = raw_table;
+        TCPIP_OWNER_MODULE_BASIC_INFO *info;
+        MIB_TCPROW_OWNER_MODULE entry;
+        WCHAR mod[MAX_PATH], *name;
+        DWORD sz, expected, name_len;
+        BYTE buf[MAX_PATH + sizeof(TCPIP_OWNER_MODULE_BASIC_INFO) + 128];
+        HANDLE process;
+        char *p;
+
+        info = (TCPIP_OWNER_MODULE_BASIC_INFO *)buf;
+
+        mod[0] = 0;
+        GetModuleFileNameW( NULL, mod, ARRAY_SIZE(mod) );
+        name = wcsrchr( mod, '\\' );
+        ok( !!name, "could not get name, mod %s.\n", debugstr_w(mod) );
+        ++name;
+        name_len = (wcslen( name ) + 1) * sizeof(WCHAR);
+        expected = sizeof(TCPIP_OWNER_MODULE_BASIC_INFO) + name_len + (wcslen( mod ) + 1) * sizeof(WCHAR);
+        memset( &entry, 0, sizeof(entry) );
+
+        sz = sizeof(buf);
+        ret = GetOwnerModuleFromTcpEntry( NULL, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( ret == ERROR_INVALID_PARAMETER, "got %lu.\n", ret );
+
+        entry.dwOwningPid = 0;
+        ret = GetOwnerModuleFromTcpEntry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( ret == ERROR_NOT_FOUND, "got %lu.\n", ret );
+
+        entry.dwOwningPid = 0xffff;
+        ret = GetOwnerModuleFromTcpEntry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( ret == ERROR_INVALID_PARAMETER, "got %lu.\n", ret );
+
+        entry.dwOwningPid = GetCurrentProcessId();
+        sz = 1;
+        ret = GetOwnerModuleFromTcpEntry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( ret == ERROR_INSUFFICIENT_BUFFER, "got %lu.\n", ret );
+        ok( sz == expected, "got %lu, expected %lu.\n", sz, expected );
+
+        ret = GetOwnerModuleFromTcpEntry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( !ret, "got %lu.\n", ret );
+
+        p = (char *)(info + 1);
+        ok( (char *)info->pModuleName == p, "got %p, %p.\n", info, p );
+        p += name_len;
+        ok( (char *)info->pModulePath == p, "got %p, %p.\n", info, p);
+        ok( !wcscmp( info->pModuleName, name ), "got %s, %s.\n", debugstr_w(info->pModuleName), debugstr_w(name) );
+
+        found_it = FALSE;
+        for (i = 0; i < t->dwNumEntries; ++i)
+        {
+            info = (TCPIP_OWNER_MODULE_BASIC_INFO *)buf;
+
+            memset(buf, 0xcc, sizeof(buf));
+            sz = sizeof(buf) + 1;
+            ret = GetOwnerModuleFromTcpEntry( &t->table[i], TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+            ok(sz == sizeof(buf) + 1, "got %ld.\n", sz);
+            ok( !ret || ret == ERROR_NOT_FOUND || ret == ERROR_ACCESS_DENIED
+                || ret == ERROR_MOD_NOT_FOUND,
+                /* This is weird, newer Windows seems to return non-zero OwningModuleInfo[0] sometimes but then
+                 * GetOwnerModuleFromTcp6Entry always fails with ERROR_MOD_NOT_FOUND, while that succeeds for the
+                 * same PID in MIB_TCPTABLE_OWNER_MODULE and zeroed OwningModuleInfo */
+                "got %#lx, pid %ld, info %#I64x.\n", ret, t->table[i].dwOwningPid, t->table[i].OwningModuleInfo[0] );
+
+            memset( &entry, 0, sizeof(entry) );
+            entry.dwOwningPid = t->table[i].dwOwningPid;
+            if (entry.dwOwningPid == GetCurrentProcessId())
+                found_it = TRUE;
+            ret = GetOwnerModuleFromTcpEntry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+            process = OpenProcess( PROCESS_QUERY_LIMITED_INFORMATION, FALSE, entry.dwOwningPid );
+            ok( !ret || ret == ERROR_NOT_FOUND || ret == ERROR_ACCESS_DENIED,
+                "got %#lx, pid %ld, info %#I64x.\n", ret, t->table[i].dwOwningPid, t->table[i].OwningModuleInfo[0] );
+            if (!ret && !wcscmp( info->pModuleName, L"System" ))
+            {
+                ok( !wcscmp( info->pModulePath, L"System" ), "got %s.\n", debugstr_w(info->pModulePath) );
+            }
+            else if (ret == ERROR_ACCESS_DENIED)
+            {
+                process = OpenProcess( PROCESS_QUERY_LIMITED_INFORMATION, FALSE, entry.dwOwningPid );
+                ok( !process && GetLastError() == ERROR_ACCESS_DENIED, "got process %p, err %lu.\n", process, GetLastError() );
+            }
+        }
+        ok( found_it, "no table entry for socket\n" );
+    }
+    else
+    {
+        MIB_TCP6TABLE_OWNER_MODULE *t = raw_table;
+        TCPIP_OWNER_MODULE_BASIC_INFO *info;
+        MIB_TCP6ROW_OWNER_MODULE entry;
+        WCHAR mod[MAX_PATH], *name;
+        DWORD sz, expected, name_len;
+        BYTE buf[MAX_PATH + sizeof(TCPIP_OWNER_MODULE_BASIC_INFO) + 128];
+        HANDLE process;
+        char *p;
+
+        info = (TCPIP_OWNER_MODULE_BASIC_INFO *)buf;
+
+        mod[0] = 0;
+        GetModuleFileNameW( NULL, mod, ARRAY_SIZE(mod) );
+        name = wcsrchr( mod, '\\' );
+        ok( !!name, "could not get name, mod %s.\n", debugstr_w(mod) );
+        ++name;
+        name_len = (wcslen( name ) + 1) * sizeof(WCHAR);
+        expected = sizeof(TCPIP_OWNER_MODULE_BASIC_INFO) + name_len + (wcslen( mod ) + 1) * sizeof(WCHAR);
+        memset( &entry, 0, sizeof(entry) );
+
+        sz = sizeof(buf);
+        ret = GetOwnerModuleFromTcp6Entry( NULL, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( ret == ERROR_INVALID_PARAMETER, "got %lu.\n", ret );
+
+        entry.dwOwningPid = 0;
+        ret = GetOwnerModuleFromTcp6Entry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( ret == ERROR_NOT_FOUND, "got %lu.\n", ret );
+
+        entry.dwOwningPid = 0xffff;
+        ret = GetOwnerModuleFromTcp6Entry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( ret == ERROR_INVALID_PARAMETER, "got %lu.\n", ret );
+
+        entry.dwOwningPid = GetCurrentProcessId();
+        sz = 1;
+        ret = GetOwnerModuleFromTcp6Entry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( ret == ERROR_INSUFFICIENT_BUFFER, "got %lu.\n", ret );
+        ok( sz == expected, "got %lu, expected %lu.\n", sz, expected );
+
+        ret = GetOwnerModuleFromTcp6Entry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+        ok( !ret, "got %lu.\n", ret );
+
+        p = (char *)(info + 1);
+        ok( (char *)info->pModuleName == p, "got %p, %p.\n", info, p );
+        p += name_len;
+        ok( (char *)info->pModulePath == p, "got %p, %p.\n", info, p);
+        ok( !wcscmp( info->pModuleName, name ), "got %s, %s.\n", debugstr_w(info->pModuleName), debugstr_w(name) );
+
+        found_it = FALSE;
+        for (i = 0; i < t->dwNumEntries; ++i)
+        {
+            info = (TCPIP_OWNER_MODULE_BASIC_INFO *)buf;
+
+            memset(buf, 0xcc, sizeof(buf));
+            sz = sizeof(buf) + 1;
+            ret = GetOwnerModuleFromTcp6Entry( &t->table[i], TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+            ok(sz == sizeof(buf) + 1, "got %ld.\n", sz);
+            ok( !ret || ret == ERROR_NOT_FOUND || ret == ERROR_ACCESS_DENIED
+                || ret == ERROR_MOD_NOT_FOUND,
+                /* This is weird, newer Windows seems to return non-zero OwningModuleInfo[0] sometimes but then
+                 * GetOwnerModuleFromTcp6Entry always fails with ERROR_MOD_NOT_FOUND, while that succeeds for the
+                 * same PID in MIB_TCPTABLE_OWNER_MODULE and zeroed OwningModuleInfo */
+                "got %#lx, pid %ld, info %#I64x.\n", ret, t->table[i].dwOwningPid, t->table[i].OwningModuleInfo[0] );
+
+            memset( &entry, 0, sizeof(entry) );
+            entry.dwOwningPid = t->table[i].dwOwningPid;
+            if (entry.dwOwningPid == GetCurrentProcessId())
+                found_it = TRUE;
+            ret = GetOwnerModuleFromTcp6Entry( &entry, TCPIP_OWNER_MODULE_INFO_BASIC, buf, &sz );
+            process = OpenProcess( PROCESS_QUERY_LIMITED_INFORMATION, FALSE, entry.dwOwningPid );
+            ok( !ret || ret == ERROR_NOT_FOUND || ret == ERROR_ACCESS_DENIED,
+                "got %#lx, pid %ld, info %#I64x.\n", ret, t->table[i].dwOwningPid, t->table[i].OwningModuleInfo[0] );
+            if (!ret && !wcscmp( info->pModuleName, L"System" ))
+            {
+                ok( !wcscmp( info->pModulePath, L"System" ), "got %s.\n", debugstr_w(info->pModulePath) );
+            }
+            else if (ret == ERROR_ACCESS_DENIED)
+            {
+                process = OpenProcess( PROCESS_QUERY_LIMITED_INFORMATION, FALSE, entry.dwOwningPid );
+                ok( !process && GetLastError() == ERROR_ACCESS_DENIED, "got process %p, err %lu.\n", process, GetLastError() );
             }
         }
         ok( found_it, "no table entry for socket\n" );
@@ -3505,13 +3686,13 @@
 
     DWORD ret, index, link_local_index = ~0u, loopback_index = ~0u, default_route_index = ~0u;
     struct sockaddr_in6 dst6, src6, best6, link_local_addr6, global_addr6;
-    struct sockaddr_in src4, global_addr4;
+    struct sockaddr_in dst4, src4, global_addr4;
     NET_LUID link_local_luid = { 0 }, luid;
     MIB_UNICASTIPADDRESS_ROW uni_row;
     MIB_IPFORWARD_ROW2 fwd_row, *r;
     MIB_IPFORWARD_TABLE2 *table;
     char s[256], s2[256];
-    SOCKADDR_INET best4, dst4;
+    SOCKADDR_INET best4;
     unsigned int i;
 
     memset( &dst6, 0, sizeof(src6) );
@@ -3890,8 +4071,8 @@
 
     /* Test with ipv4 */
     memset( &dst4, 0xcc, sizeof(dst4) );
-    dst4.Ipv4.sin_family = AF_INET;
-    ret = inet_pton( AF_INET, "25.25.0.0", &dst4.Ipv4.sin_addr);
+    dst4.sin_family = AF_INET;
+    ret = inet_pton( AF_INET, "25.25.0.0", &dst4.sin_addr);
     ok(ret, "got error %u.\n", WSAGetLastError());
 
     ret = GetBestInterfaceEx( (struct sockaddr *)&dst4, &index );
@@ -3905,7 +4086,7 @@
 
     memset( &fwd_row, 0xcc, sizeof(fwd_row) );
     memset( &best4, 0xcc, sizeof(best4) );
-    ret = GetBestRoute2( NULL, default_route_index, NULL, &dst4, 0, &fwd_row, &best4 );
+    ret = GetBestRoute2( NULL, default_route_index, NULL, (SOCKADDR_INET *)&dst4, 0, &fwd_row, (SOCKADDR_INET *)&best4 );
     ok( !ret, "got error %lu.\n", ret );
     ok( fwd_row.InterfaceIndex == default_route_index, "got %lu, expected %lu.\n", fwd_row.InterfaceIndex,
         default_route_index );
@@ -3926,14 +4107,14 @@
     ret = inet_pton( AF_INET, "127.0.0.1", &src4.sin_addr );
     ok(ret, "got error %u.\n", WSAGetLastError());
     memset( &best4, 0xcc, sizeof(best4) );
-    ret = GetBestRoute2( NULL, 0, (SOCKADDR_INET *)&src4, &dst4, 0, &fwd_row, &best4 );
+    ret = GetBestRoute2( NULL, 0, (SOCKADDR_INET *)&src4, (SOCKADDR_INET *)&dst4, 0, &fwd_row, (SOCKADDR_INET *)&best4 );
     ok( ret == ERROR_NETWORK_UNREACHABLE, "got error %lu.\n", ret );
     ok( !best4.Ipv4.sin_family, "got %u.\n", best4.Ipv4.sin_family );
     ok( !best4.Ipv6.sin6_addr.u.Word[7], "got %#x.\n", best4.Ipv6.sin6_addr.u.Word[7] );
 
     src4 = global_addr4;
-    ret = GetBestRoute2( NULL, loopback_index, (SOCKADDR_INET *)&src4, &dst4, 0, &fwd_row,
-                         &best4 );
+    ret = GetBestRoute2( NULL, loopback_index, (SOCKADDR_INET *)&src4, (SOCKADDR_INET *)&dst4, 0, &fwd_row,
+                         (SOCKADDR_INET *)&best4 );
     ok( !ret, "got error %lu.\n", ret );
     ok( fwd_row.InterfaceIndex == default_route_index, "got %lu, expected %lu.\n", fwd_row.InterfaceIndex,
         link_local_index );
@@ -3942,29 +4123,29 @@
 
     memset( &src4, 0xcc, sizeof(src4) );
     memset( &fwd_row, 0xcc, sizeof(fwd_row) );
-    ret = GetBestRoute2( NULL, default_route_index, (SOCKADDR_INET *)&src4, &dst4, 0, &fwd_row,
-                         &best4 );
+    ret = GetBestRoute2( NULL, default_route_index, (SOCKADDR_INET *)&src4, (SOCKADDR_INET *)&dst4, 0, &fwd_row,
+                         (SOCKADDR_INET *)&best4 );
     ok( !ret, "got error %lu.\n", ret );
     ok( fwd_row.InterfaceIndex == default_route_index, "got %lu, expected %lu.\n", fwd_row.InterfaceIndex,
         link_local_index );
 
     src4 = best4.Ipv4;
-    ret = GetBestRoute2( NULL, default_route_index, (SOCKADDR_INET *)&src4, &dst4, 0, &fwd_row,
-                         &best4 );
+    ret = GetBestRoute2( NULL, default_route_index, (SOCKADDR_INET *)&src4, (SOCKADDR_INET *)&dst4, 0, &fwd_row,
+                         (SOCKADDR_INET *)&best4 );
     ok( !ret, "got error %lu.\n", ret );
     ok( fwd_row.InterfaceIndex == default_route_index, "got %lu, expected %lu.\n", fwd_row.InterfaceIndex,
         link_local_index );
     ok( fwd_row.NextHop.Ipv4.sin_addr.s_addr, "expected specified next hop.\n" );
 
-    ret = GetBestRoute2( NULL, loopback_index, NULL, &dst4, 0, &fwd_row, &best4 );
+    ret = GetBestRoute2( NULL, loopback_index, NULL, (SOCKADDR_INET *)&dst4, 0, &fwd_row, (SOCKADDR_INET *)&best4 );
     ok( ret == ERROR_NETWORK_UNREACHABLE, "got error %lu.\n", ret );
 
-    ret = inet_pton( AF_INET, "127.0.0.1", &dst4.Ipv4.sin_addr);
+    ret = inet_pton( AF_INET, "127.0.0.1", &dst4.sin_addr);
     ok(ret, "got error %u.\n", WSAGetLastError());
-    ret = GetBestRoute2( NULL, default_route_index, NULL, &dst4, 0, &fwd_row, &best4 );
+    ret = GetBestRoute2( NULL, default_route_index, NULL, (SOCKADDR_INET *)&dst4, 0, &fwd_row, (SOCKADDR_INET *)&best4 );
     ok( ret == ERROR_INVALID_PARAMETER, "got error %lu.\n", ret );
 
-    ret = GetBestRoute2( NULL, 0, NULL, &dst4, 0, &fwd_row, &best4 );
+    ret = GetBestRoute2( NULL, 0, NULL, (SOCKADDR_INET *)&dst4, 0, &fwd_row, (SOCKADDR_INET *)&best4 );
     ok( !ret, "got error %lu.\n", ret );
     ok( !fwd_row.NextHop.Ipv4.sin_addr.s_addr, "expected unspecified next hop.\n" );
 
