diff --git a/dlls/wineps.drv/type1afm.c b/dlls/wineps.drv/type1afm.c
--- a/dlls/wineps.drv/type1afm.c
+++ b/dlls/wineps.drv/type1afm.c
@@ -540,13 +540,81 @@
     if (retval == FALSE || found == FALSE)
     	goto cleanup_afm;
 
+    retval = ReadFloat(file, buffer, bufsize, "UnderlinePosition",
+    	    &(afm->UnderlinePosition), &found);
+    if (retval == FALSE || found == FALSE)
+    	goto cleanup_afm;
+
+    retval = ReadFloat(file, buffer, bufsize, "UnderlineThickness",
+    	    &(afm->UnderlineThickness), &found);
+    if (retval == FALSE || found == FALSE)
+    	goto cleanup_afm;
+
+    retval = ReadFloat(file, buffer, bufsize, "Ascender",    	/* optional */
+    	    &(afm->Ascender), &found);
+    if (retval == FALSE)
+    	goto cleanup_afm;
+
+    retval = ReadFloat(file, buffer, bufsize, "Descender",   	/* optional */
+    	    &(afm->Descender), &found);
+    if (retval == FALSE)
+    	goto cleanup_afm;
+
     afm->WinMetrics.usUnitsPerEm = 1000;
+    afm->WinMetrics.sTypoAscender = (SHORT)Round(afm->Ascender);
+    afm->WinMetrics.sTypoDescender = (SHORT)Round(afm->Descender);
+
+    if (afm->WinMetrics.sTypoAscender == 0)
+    	afm->WinMetrics.sTypoAscender = (SHORT)Round(afm->FontBBox.ury);
+
+    if (afm->WinMetrics.sTypoDescender == 0)
+    	afm->WinMetrics.sTypoDescender = (SHORT)Round(afm->FontBBox.lly);
+
+    afm->WinMetrics.sTypoLineGap = 1200 -
+    	    (afm->WinMetrics.sTypoAscender - afm->WinMetrics.sTypoDescender);
+    if (afm->WinMetrics.sTypoLineGap < 0)
+    	afm->WinMetrics.sTypoLineGap = 0;
+
     return TRUE;
 
     cleanup_afm:    	    	    	/* handle fatal or non-fatal errors */
     	HeapFree(PSDRV_Heap, 0, afm);
 	*p_afm = NULL;
 	return retval;
+}
+
+/*******************************************************************************
+ *  ParseC
+ *
+ *  Fatal error:    	return FALSE (none defined)
+ *
+ *  Non-fatal error:	leave metrics->C set to INT_MAX
+ *
+ */
+static BOOL ParseC(LPSTR sz, OLD_AFMMETRICS *metrics)
+{
+    int     base = 10;
+    long    l;
+    CHAR    *cp, *end_ptr;
+
+    cp = sz + 1;
+
+    if (*cp == 'H')
+    {
+    	base = 16;
+	++cp;
+    }
+
+    errno = 0;
+    l = strtol(cp, &end_ptr, base);
+    if (end_ptr == cp || errno != 0 || l > INT_MAX || l < INT_MIN)
+    {
+    	WARN("Error parsing character code '%s'\n", sz);
+	return TRUE;
+    }
+
+    metrics->C = (INT)l;
+    return TRUE;
 }
 
 /*******************************************************************************
@@ -659,14 +727,15 @@
  */
 static int __cdecl ug_name_cmp(const void *a, const void *b)
 {
-    return strcmp(((const UNICODEGLYPH *)a)->name,
-            ((const UNICODEGLYPH *)b)->name);
+    return strcmp(((const UNICODEGLYPH *)a)->name->sz,
+            ((const UNICODEGLYPH *)b)->name->sz);
 }
 
 static BOOL ParseN(LPSTR sz, OLD_AFMMETRICS *metrics)
 {
     CHAR    save, *cp, *end_ptr;
     UNICODEGLYPH ug, *pug;
+    GLYPHNAME gn;
 
     cp = sz + 1;
 
@@ -687,16 +756,19 @@
     save = *end_ptr;
     *end_ptr = '\0';
 
-    ug.name = cp;
+    ug.name = &gn;
+    gn.sz = cp;
     pug = bsearch(&ug, PSDRV_AGLbyName, PSDRV_AGLbyNameSize,
             sizeof(ug), ug_name_cmp);
     if (!pug)
     {
         FIXME("unsupported glyph name: %s\n", cp);
+        metrics->N = NULL;
         metrics->UV = -1;
     }
     else
     {
+        metrics->N = pug->name;
         metrics->UV = pug->UV;
     }
 
@@ -713,9 +785,12 @@
  */
 static const OLD_AFMMETRICS badmetrics =
 {
-    -1,		    	    	    	    	    /* UV */
+    INT_MAX,	    	    	    	    	    /* C */
+    INT_MAX,	    	    	    	    	    /* UV */
     FLT_MAX,	    	    	    	    	    /* WX */
+    NULL,   	    	    	    	    	    /* N */
     { FLT_MAX, FLT_MAX, FLT_MAX, FLT_MAX }, 	    /* B */
+    NULL    	    	    	    	    	    /* L */
 };
 
 static BOOL ParseCharMetrics(LPSTR buffer, INT len, OLD_AFMMETRICS *metrics)
@@ -731,6 +806,10 @@
 
 	switch(*cp)
 	{
+	    case 'C':	if (ParseC(cp, metrics) == FALSE)
+	    	    	    return FALSE;
+	    	    	break;
+
 	    case 'W':	if (ParseW(cp, metrics) == FALSE)
 	    	    	    return FALSE;
 	    	    	break;
@@ -754,7 +833,8 @@
 	++cp;
     }
 
-    if (metrics->WX == FLT_MAX || metrics->UV == -1 || metrics->B.ury == FLT_MAX)
+    if (metrics->C == INT_MAX || metrics->WX == FLT_MAX || metrics->N == NULL ||
+    	    metrics->B.ury == FLT_MAX)
     {
     	*metrics = badmetrics;
 	return TRUE;
@@ -892,7 +972,7 @@
 	if(len > 0)
 	{
 	    retval = ParseCharMetrics(buffer, len, old_metrics + i);
-	    if (retval == FALSE || old_metrics[i].UV == -1)
+	    if (retval == FALSE || old_metrics[i].C == INT_MAX)
 	    	goto cleanup_old_metrics;
 
 	    continue;
@@ -928,8 +1008,10 @@
 
     for (i = 0; i < afm->NumofMetrics; ++i, ++metrics, ++encoded_metrics)
     {
+    	metrics->C = encoded_metrics->C;
 	metrics->UV = encoded_metrics->UV;
 	metrics->WX = encoded_metrics->WX;
+	metrics->N = encoded_metrics->N;
     }
 
     HeapFree(PSDRV_Heap, 0, old_metrics);
@@ -956,7 +1038,7 @@
     CHAR    	buffer[258];    	/* allow for <cr>, <lf>, and <nul> */
     AFM     	*afm;
     AFMMETRICS	*metrics;
-    WCHAR *family_name, *encoding_scheme;
+    WCHAR *full_name, *family_name, *encoding_scheme;
     char *font_name;
     BOOL retval, added;
 
@@ -968,10 +1050,14 @@
     if (retval == FALSE || font_name == NULL)
     	goto cleanup_afm;
 
+    retval = ReadStringW(file, buffer, sizeof(buffer), "FullName", &full_name);
+    if (retval == FALSE || full_name == NULL)
+    	goto cleanup_font_name;
+
     retval = ReadStringW(file, buffer, sizeof(buffer), "FamilyName",
     	    &family_name);
     if (retval == FALSE || family_name == NULL)
-    	goto cleanup_font_name;
+    	goto cleanup_full_name;
 
     retval = ReadStringW(file, buffer, sizeof(buffer), "EncodingScheme",
     	    &encoding_scheme);
@@ -979,6 +1065,7 @@
     	goto cleanup_family_name;
 
     afm->FontName = font_name;
+    afm->FullName = full_name;
     afm->FamilyName = family_name;
     afm->EncodingScheme = encoding_scheme;
 
@@ -998,6 +1085,8 @@
     	HeapFree(PSDRV_Heap, 0, encoding_scheme);
     cleanup_family_name:
     	HeapFree(PSDRV_Heap, 0, family_name);
+    cleanup_full_name:
+    	HeapFree(PSDRV_Heap, 0, full_name);
     cleanup_font_name:
     	HeapFree(PSDRV_Heap, 0, font_name);
     cleanup_afm:
@@ -1039,17 +1128,25 @@
  *  Reads all Type 1 AFM files in a directory.
  *
  */
-static BOOL ReadAFMDir(const WCHAR *dirname)
-{
+static BOOL ReadAFMDir(LPCSTR dirname)
+{
+    WCHAR *path = wine_get_dos_file_name( dirname );
     struct _wfinddata_t data;
     intptr_t handle;
     WCHAR *p, *filename;
     BOOL ret = TRUE;
-    DWORD size = 8 + wcslen(dirname) + 2 + 1;
-
-    if (!(filename = malloc( size * sizeof(WCHAR) ))) return FALSE;
-    swprintf( filename, size, L"\\\\?\\unix%s\\*", dirname );
-    p = filename + wcslen(filename) - 1;
+
+    if (!path) return TRUE;
+    if (!(filename = HeapAlloc( GetProcessHeap(), 0, lstrlenW(path) + 256 )))
+    {
+        HeapFree( GetProcessHeap(), 0, path );
+        return FALSE;
+    }
+    lstrcpyW( filename, path );
+    HeapFree( GetProcessHeap(), 0, path );
+    p = filename + lstrlenW(filename);
+    *p++ = '\\';
+    lstrcpyW( p, L"*" );
 
     handle = _wfindfirst( filename, &data );
     if (handle != -1)
@@ -1063,7 +1160,8 @@
         } while (!_wfindnext( handle, &data ));
     }
     _findclose( handle );
-    free( filename );
+
+    HeapFree( GetProcessHeap(), 0, filename );
     return ret;
 }
 
@@ -1082,7 +1180,8 @@
 {
     HKEY hkey;
     DWORD len;
-    WCHAR *value, *ptr, *next;
+    LPWSTR valueW;
+    LPSTR valueA, ptr;
 
     /* @@ Wine registry key: HKCU\Software\Wine\Fonts */
     if (RegOpenKeyA(HKEY_CURRENT_USER, "Software\\Wine\\Fonts", &hkey) != ERROR_SUCCESS)
@@ -1091,21 +1190,28 @@
     if (RegQueryValueExW( hkey, L"AFMPath", NULL, NULL, NULL, &len ) == ERROR_SUCCESS)
     {
         len += sizeof(WCHAR);
-        value = HeapAlloc( PSDRV_Heap, 0, len );
-        if (RegQueryValueExW( hkey, L"AFMPath", NULL, NULL, (BYTE *)value, &len ) == ERROR_SUCCESS)
+        valueW = HeapAlloc( PSDRV_Heap, 0, len );
+        if (RegQueryValueExW( hkey, L"AFMPath", NULL, NULL, (BYTE *)valueW, &len ) == ERROR_SUCCESS)
         {
-            for (ptr = value; ptr; ptr = next)
+            len = WideCharToMultiByte( CP_UNIXCP, 0, valueW, -1, NULL, 0, NULL, NULL );
+            valueA = HeapAlloc( PSDRV_Heap, 0, len );
+            WideCharToMultiByte( CP_UNIXCP, 0, valueW, -1, valueA, len, NULL, NULL );
+            TRACE( "got AFM font path %s\n", debugstr_a(valueA) );
+            ptr = valueA;
+            while (ptr)
             {
-                next = wcschr( ptr, ':' );
+                LPSTR next = strchr( ptr, ':' );
                 if (next) *next++ = 0;
                 if (!ReadAFMDir( ptr ))
                 {
                     RegCloseKey(hkey);
                     return FALSE;
                 }
+                ptr = next;
             }
+            HeapFree( PSDRV_Heap, 0, valueA );
         }
-        HeapFree( PSDRV_Heap, 0, value );
+        HeapFree( PSDRV_Heap, 0, valueW );
     }
 
     RegCloseKey(hkey);
