diff --git a/dlls/msxml3/saxreader.c b/dlls/msxml3/saxreader.c
--- a/dlls/msxml3/saxreader.c
+++ b/dlls/msxml3/saxreader.c
@@ -63,6 +63,45 @@
     LexicalHandlerParEntities    = 1 << 15
 } saxreader_feature;
 
+/* feature names */
+static const WCHAR FeatureExternalGeneralEntitiesW[] = {
+    'h','t','t','p',':','/','/','x','m','l','.','o','r','g','/','s','a','x','/',
+    'f','e','a','t','u','r','e','s','/','e','x','t','e','r','n','a','l','-','g','e','n','e','r','a','l',
+    '-','e','n','t','i','t','i','e','s',0
+};
+
+static const WCHAR FeatureExternalParameterEntitiesW[] = {
+    'h','t','t','p',':','/','/','x','m','l','.','o','r','g','/','s','a','x','/','f','e','a','t','u','r','e','s',
+    '/','e','x','t','e','r','n','a','l','-','p','a','r','a','m','e','t','e','r','-','e','n','t','i','t','i','e','s',0
+};
+
+static const WCHAR FeatureLexicalHandlerParEntitiesW[] = {
+    'h','t','t','p',':','/','/','x','m','l','.','o','r','g','/','s','a','x','/','f','e','a','t','u','r','e','s',
+    '/','l','e','x','i','c','a','l','-','h','a','n','d','l','e','r','/','p','a','r','a','m','e','t','e','r','-','e','n','t','i','t','i','e','s',0
+};
+
+static const WCHAR FeatureProhibitDTDW[] = {
+    'p','r','o','h','i','b','i','t','-','d','t','d',0
+};
+
+static const WCHAR FeatureNamespacesW[] = {
+    'h','t','t','p',':','/','/','x','m','l','.','o','r','g','/','s','a','x','/','f','e','a','t','u','r','e','s',
+    '/','n','a','m','e','s','p','a','c','e','s',0
+};
+
+static const WCHAR FeatureNamespacePrefixesW[] = {
+    'h','t','t','p',':','/','/','x','m','l','.','o','r','g','/','s','a','x','/','f','e','a','t','u','r','e','s',
+    '/','n','a','m','e','s','p','a','c','e','-','p','r','e','f','i','x','e','s',0
+};
+
+static const WCHAR ExhaustiveErrorsW[] = {
+    'e','x','h','a','u','s','t','i','v','e','-','e','r','r','o','r','s',0
+};
+
+static const WCHAR SchemaValidationW[] = {
+    's','c','h','e','m','a','-','v','a','l','i','d','a','t','i','o','n',0
+};
+
 struct saxreader_feature_pair
 {
     saxreader_feature feature;
@@ -70,14 +109,14 @@
 };
 
 static const struct saxreader_feature_pair saxreader_feature_map[] = {
-    { ExhaustiveErrors, L"exhaustive-errors" },
-    { ExternalGeneralEntities, L"http://xml.org/sax/features/external-general-entities" },
-    { ExternalParameterEntities, L"http://xml.org/sax/features/external-parameter-entities" },
-    { LexicalHandlerParEntities, L"http://xml.org/sax/features/lexical-handler/parameter-entities" },
-    { NamespacePrefixes, L"http://xml.org/sax/features/namespace-prefixes" },
-    { Namespaces, L"http://xml.org/sax/features/namespaces" },
-    { ProhibitDTD, L"prohibit-dtd" },
-    { SchemaValidation, L"schema-validation" },
+    { ExhaustiveErrors, ExhaustiveErrorsW },
+    { ExternalGeneralEntities, FeatureExternalGeneralEntitiesW },
+    { ExternalParameterEntities, FeatureExternalParameterEntitiesW },
+    { LexicalHandlerParEntities, FeatureLexicalHandlerParEntitiesW },
+    { NamespacePrefixes, FeatureNamespacePrefixesW },
+    { Namespaces, FeatureNamespacesW },
+    { ProhibitDTD, FeatureProhibitDTDW },
+    { SchemaValidation, SchemaValidationW },
 };
 
 static saxreader_feature get_saxreader_feature(const WCHAR *name)
@@ -331,9 +370,50 @@
 }
 
 /* property names */
-static const WCHAR PropertyDeclHandlerW[] = L"http://xml.org/sax/properties/declaration-handler";
-static const WCHAR PropertyDomNodeW[] = L"http://xml.org/sax/properties/dom-node";
-static const WCHAR PropertyLexicalHandlerW[] = L"http://xml.org/sax/properties/lexical-handler";
+static const WCHAR PropertyCharsetW[] = {
+    'c','h','a','r','s','e','t',0
+};
+static const WCHAR PropertyXmlDeclVersionW[] = {
+    'x','m','l','d','e','c','l','-','v','e','r','s','i','o','n',0
+};
+static const WCHAR PropertyDeclHandlerW[] = {
+    'h','t','t','p',':','/','/','x','m','l','.','o','r','g','/',
+    's','a','x','/','p','r','o','p','e','r','t','i','e','s','/',
+    'd','e','c','l','a','r','a','t','i','o','n',
+    '-','h','a','n','d','l','e','r',0
+};
+static const WCHAR PropertyDomNodeW[] = {
+    'h','t','t','p',':','/','/','x','m','l','.','o','r','g','/',
+    's','a','x','/','p','r','o','p','e','r','t','i','e','s','/',
+    'd','o','m','-','n','o','d','e',0
+};
+static const WCHAR PropertyInputSourceW[] = {
+    'i','n','p','u','t','-','s','o','u','r','c','e',0
+};
+static const WCHAR PropertyLexicalHandlerW[] = {
+    'h','t','t','p',':','/','/','x','m','l','.','o','r','g','/',
+    's','a','x','/','p','r','o','p','e','r','t','i','e','s','/',
+    'l','e','x','i','c','a','l','-','h','a','n','d','l','e','r',0
+};
+static const WCHAR PropertyMaxElementDepthW[] = {
+    'm','a','x','-','e','l','e','m','e','n','t','-','d','e','p','t','h',0
+};
+static const WCHAR PropertyMaxXMLSizeW[] = {
+    'm','a','x','-','x','m','l','-','s','i','z','e',0
+};
+static const WCHAR PropertySchemaDeclHandlerW[] = {
+    's','c','h','e','m','a','-','d','e','c','l','a','r','a','t','i','o','n','-',
+    'h','a','n','d','l','e','r',0
+};
+static const WCHAR PropertyXMLDeclEncodingW[] = {
+    'x','m','l','d','e','c','l','-','e','n','c','o','d','i','n','g',0
+};
+static const WCHAR PropertyXMLDeclStandaloneW[] = {
+    'x','m','l','d','e','c','l','-','s','t','a','n','d','a','l','o','n','e',0
+};
+static const WCHAR PropertyXMLDeclVersionW[] = {
+    'x','m','l','d','e','c','l','-','v','e','r','s','i','o','n',0
+};
 
 static inline HRESULT set_feature_value(saxreader *reader, saxreader_feature feature, VARIANT_BOOL value)
 {
@@ -1308,7 +1388,7 @@
    escaped to '&#38;'. This function takes care of ampersands only. */
 static BSTR saxreader_get_unescaped_value(const xmlChar *buf, int len)
 {
-    static const WCHAR ampescW[] = L"&#38;";
+    static const WCHAR ampescW[] = {'&','#','3','8',';',0};
     WCHAR *dest, *ptrW, *str;
     DWORD str_len;
     BSTR bstr;
@@ -1368,6 +1448,7 @@
         int nb_attributes, const xmlChar **xmlAttributes)
 {
     static const xmlChar xmlns[] = "xmlns";
+    static const WCHAR xmlnsW[] = { 'x','m','l','n','s',0 };
 
     struct _attributes *attrs;
     int i;
@@ -1409,7 +1490,7 @@
 
         SysFreeString(attrs[nb_attributes+i].szQName);
         if(!xmlNamespaces[2*i])
-            attrs[nb_attributes+i].szQName = SysAllocString(L"xmlns");
+            attrs[nb_attributes+i].szQName = SysAllocString(xmlnsW);
         else
             attrs[nb_attributes+i].szQName = QName_from_xmlChar(xmlns, xmlNamespaces[2*i]);
     }
@@ -2000,71 +2081,6 @@
         format_error_message_from_id(locator, hr);
 }
 
-static void libxml_pi(void *ctx, const xmlChar *_target, const xmlChar *_data)
-{
-    saxlocator *locator = ctx;
-    struct saxcontenthandler_iface *handler = saxreader_get_contenthandler(locator->saxreader);
-    BSTR target, data;
-    HRESULT hr = S_OK;
-
-    update_position(locator, FALSE);
-    if (!saxreader_has_handler(locator, SAXContentHandler)) return;
-
-    target = pooled_bstr_from_xmlChar(&locator->saxreader->pool, _target);
-    data = pooled_bstr_from_xmlChar(&locator->saxreader->pool, _data);
-
-    if (locator->vbInterface)
-        hr = IVBSAXContentHandler_processingInstruction(handler->vbhandler, &target, &data);
-    else
-        hr = ISAXContentHandler_processingInstruction(handler->handler, target, SysStringLen(target),
-                data, SysStringLen(data));
-
-    if (FAILED(hr))
-        format_error_message_from_id(locator, hr);
-}
-
-static void libxml_internalsubset(void *ctx, const xmlChar *_name, const xmlChar *external_id, const xmlChar *system_id)
-{
-    saxlocator *locator = ctx;
-    struct saxlexicalhandler_iface *lexical = saxreader_get_lexicalhandler(locator->saxreader);
-    BSTR name, pubid, sysid;
-    HRESULT hr;
-
-    update_position(locator, FALSE);
-    if (!saxreader_has_handler(locator, SAXLexicalHandler)) return;
-
-    name = pooled_bstr_from_xmlChar(&locator->saxreader->pool, _name);
-    pubid = pooled_bstr_from_xmlChar(&locator->saxreader->pool, external_id);
-    sysid = pooled_bstr_from_xmlChar(&locator->saxreader->pool, system_id);
-
-    if (locator->vbInterface)
-        hr = IVBSAXLexicalHandler_startDTD(lexical->vbhandler, &name, &pubid, &sysid);
-    else
-        hr = ISAXLexicalHandler_startDTD(lexical->handler, name, SysStringLen(name),
-                pubid, SysStringLen(pubid), sysid, SysStringLen(sysid));
-
-    if (FAILED(hr))
-        format_error_message_from_id(locator, hr);
-}
-
-static void libxml_externalsubset(void *ctx, const xmlChar *name, const xmlChar *external_id, const xmlChar *system_id)
-{
-    saxlocator *locator = ctx;
-    struct saxlexicalhandler_iface *lexical = saxreader_get_lexicalhandler(locator->saxreader);
-    HRESULT hr = S_OK;
-
-    update_position(locator, FALSE);
-    if (!saxreader_has_handler(locator, SAXLexicalHandler)) return;
-
-    if (locator->vbInterface)
-        hr = IVBSAXLexicalHandler_endDTD(lexical->vbhandler);
-    else
-        hr = ISAXLexicalHandler_endDTD(lexical->handler);
-
-    if (FAILED(hr))
-        format_error_message_from_id(locator, hr);
-}
-
 static xmlParserInputPtr libxmlresolveentity(void *ctx, const xmlChar *publicid, const xmlChar *systemid)
 {
     FIXME("entity resolving not implemented, %s, %s\n", publicid, systemid);
@@ -2415,6 +2431,9 @@
 
 static HRESULT SAXLocator_create(saxreader *reader, saxlocator **ppsaxlocator, BOOL vbInterface)
 {
+    static const WCHAR w3xmlns[] = { 'h','t','t','p',':','/','/', 'w','w','w','.','w','3','.',
+        'o','r','g','/','2','0','0','0','/','x','m','l','n','s','/',0 };
+
     saxlocator *locator;
 
     locator = malloc(sizeof(*locator));
@@ -2438,7 +2457,7 @@
     locator->column = 0;
     locator->ret = S_OK;
     if (locator->saxreader->version >= MSXML6)
-        locator->namespaceUri = SysAllocString(L"http://www.w3.org/2000/xmlns/");
+        locator->namespaceUri = SysAllocString(w3xmlns);
     else
         locator->namespaceUri = SysAllocStringLen(NULL, 0);
     if(!locator->namespaceUri)
@@ -2781,14 +2800,14 @@
     if(!memcmp(prop, PropertyLexicalHandlerW, sizeof(PropertyLexicalHandlerW)))
         return saxreader_put_handler_from_variant(This, SAXLexicalHandler, v, vbInterface);
 
-    if(!memcmp(prop, L"max-xml-size", sizeof(L"max-xml-size")))
+    if(!memcmp(prop, PropertyMaxXMLSizeW, sizeof(PropertyMaxXMLSizeW)))
     {
         if (V_VT(v) == VT_I4 && V_I4(v) == 0) return S_OK;
         FIXME("(%p)->(%s): max-xml-size unsupported\n", This, debugstr_variant(v));
         return E_NOTIMPL;
     }
 
-    if(!memcmp(prop, L"max-element-depth", sizeof(L"max-element-depth")))
+    if(!memcmp(prop, PropertyMaxElementDepthW, sizeof(PropertyMaxElementDepthW)))
     {
         if (V_VT(v) == VT_I4 && V_I4(v) == 0) return S_OK;
         FIXME("(%p)->(%s): max-element-depth unsupported\n", This, debugstr_variant(v));
@@ -2797,25 +2816,25 @@
 
     FIXME("(%p)->(%s:%s): unsupported property\n", This, debugstr_w(prop), debugstr_variant(v));
 
-    if(!memcmp(prop, L"charset", sizeof(L"charset")))
+    if(!memcmp(prop, PropertyCharsetW, sizeof(PropertyCharsetW)))
         return E_NOTIMPL;
 
     if(!memcmp(prop, PropertyDomNodeW, sizeof(PropertyDomNodeW)))
         return E_FAIL;
 
-    if(!memcmp(prop, L"input-source", sizeof(L"input-source")))
+    if(!memcmp(prop, PropertyInputSourceW, sizeof(PropertyInputSourceW)))
         return E_NOTIMPL;
 
-    if(!memcmp(prop, L"schema-declaration-handler", sizeof(L"schema-declaration-handler")))
+    if(!memcmp(prop, PropertySchemaDeclHandlerW, sizeof(PropertySchemaDeclHandlerW)))
         return E_NOTIMPL;
 
-    if(!memcmp(prop, L"xmldecl-encoding", sizeof(L"xmldecl-encoding")))
+    if(!memcmp(prop, PropertyXMLDeclEncodingW, sizeof(PropertyXMLDeclEncodingW)))
         return E_FAIL;
 
-    if(!memcmp(prop, L"xmldecl-standalone", sizeof(L"xmldecl-standalone")))
+    if(!memcmp(prop, PropertyXMLDeclStandaloneW, sizeof(PropertyXMLDeclStandaloneW)))
         return E_FAIL;
 
-    if(!memcmp(prop, L"xmldecl-version", sizeof(L"xmldecl-version")))
+    if(!memcmp(prop, PropertyXMLDeclVersionW, sizeof(PropertyXMLDeclVersionW)))
         return E_FAIL;
 
     return E_INVALIDARG;
@@ -2841,7 +2860,7 @@
         return S_OK;
     }
 
-    if (!memcmp(L"xmldecl-version", prop, sizeof(L"xmldecl-version")))
+    if (!memcmp(PropertyXmlDeclVersionW, prop, sizeof(PropertyXmlDeclVersionW)))
     {
         V_VT(value) = VT_BSTR;
         V_BSTR(value) = SysAllocString(This->xmldecl_version);
@@ -3442,9 +3461,6 @@
     reader->sax.fatalError = libxmlFatalError;
     reader->sax.cdataBlock = libxml_cdatablock;
     reader->sax.resolveEntity = libxmlresolveentity;
-    reader->sax.processingInstruction = libxml_pi;
-    reader->sax.internalSubset = libxml_internalsubset;
-    reader->sax.externalSubset = libxml_externalsubset;
 
     *ppObj = &reader->IVBSAXXMLReader_iface;
 
