diff --git a/dlls/shell32/shellpath.c b/dlls/shell32/shellpath.c
--- a/dlls/shell32/shellpath.c
+++ b/dlls/shell32/shellpath.c
@@ -679,6 +679,11 @@
 	return PathQualifyA(pszPath);
 }
 
+BOOL WINAPI PathFindOnPathExA(LPSTR,LPCSTR *,DWORD);
+BOOL WINAPI PathFindOnPathExW(LPWSTR,LPCWSTR *,DWORD);
+BOOL WINAPI PathFileExistsDefExtA(LPSTR,DWORD);
+BOOL WINAPI PathFileExistsDefExtW(LPWSTR,DWORD);
+
 static BOOL PathResolveA(char *path, const char **dirs, DWORD flags)
 {
     BOOL is_file_spec = PathIsFileSpecA(path);
@@ -2039,7 +2044,7 @@
     },
     { /* 0x6e */
         .id         = &FOLDERID_UserProgramFiles,
-        .type       = CSIDL_Type_User,
+        .type       = CSIDL_Type_Disallowed, /* FIXME */
         .category   = KF_CATEGORY_PERUSER,
         .name       = L"UserProgramFiles",
         .parent     = &FOLDERID_LocalAppData,
@@ -2663,180 +2668,6 @@
     TRACE("returning 0x%08lx (input was %s, output is %s)\n", hr,
      debugstr_w(szSrc), debugstr_w(szDest));
     return hr;
-}
-
-static char *xdg_config;
-static DWORD xdg_config_len;
-
-static BOOL WINAPI init_xdg_dirs( INIT_ONCE *once, void *param, void **context )
-{
-    const WCHAR *var, *fmt = L"\\??\\unix%s/user-dirs.dirs";
-    char *p;
-    WCHAR *name, *ptr;
-    HANDLE file;
-    DWORD len;
-
-    if (!(var = _wgetenv( L"WINE_HOST_XDG_CONFIG_HOME" )) || var[0] != '/')
-    {
-        if (!(var = _wgetenv( L"WINEHOMEDIR" ))) return TRUE;
-        fmt = L"%s/.config/user-dirs.dirs";
-    }
-    len = lstrlenW(var) + lstrlenW(fmt);
-    name = malloc( len * sizeof(WCHAR) );
-    swprintf( name, len, fmt, var );
-    name[1] = '\\';  /* change \??\ to \\?\ */
-    for (ptr = name; *ptr; ptr++) if (*ptr == '/') *ptr = '\\';
-
-    file = CreateFileW( name, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, 0 );
-    free( name );
-    if (file != INVALID_HANDLE_VALUE)
-    {
-        len = GetFileSize( file, NULL );
-        if (!(xdg_config = malloc( len + 1 ))) return TRUE;
-        if (!ReadFile( file, xdg_config, len, &xdg_config_len, NULL ))
-        {
-            free( xdg_config );
-            xdg_config = NULL;
-        }
-        else
-        {
-            for (p = xdg_config; p < xdg_config + xdg_config_len; p++) if (*p == '\n') *p = 0;
-            *p = 0;  /* append null to simplify string parsing */
-        }
-        CloseHandle( file );
-    }
-    return TRUE;
-}
-
-static char *get_xdg_path( const char *var )
-{
-    static INIT_ONCE once;
-    char *p, *ret = NULL;
-    int i;
-
-    InitOnceExecuteOnce( &once, init_xdg_dirs, NULL, NULL );
-    if (!xdg_config) return NULL;
-
-    for (p = xdg_config; p < xdg_config + xdg_config_len; p += strlen(p) + 1)
-    {
-        while (*p == ' ' || *p == '\t') p++;
-        if (strncmp( p, var, strlen(var) )) continue;
-        p += strlen(var);
-        while (*p == ' ' || *p == '\t') p++;
-        if (*p != '=') continue;
-        p++;
-        while (*p == ' ' || *p == '\t') p++;
-        if (*p != '"') continue;
-        p++;
-        if (*p != '/' && strncmp( p, "$HOME/", 6 )) continue;
-
-        if (!(ret = malloc( strlen(p) + 1 ))) break;
-        for (i = 0; *p && *p != '"'; i++, p++)
-        {
-            if (*p == '\\' && p[1]) p++;
-            ret[i] = *p;
-        }
-        ret[i] = 0;
-        if (*p != '"')
-        {
-            free( ret );
-            ret = NULL;
-        }
-        break;
-    }
-    return ret;
-}
-
-static BOOL link_folder( HANDLE mgr, const UNICODE_STRING *path, const char *link )
-{
-    struct mountmgr_shell_folder *ioctl;
-    DWORD len = sizeof(*ioctl) + path->Length + strlen(link) + 1;
-    BOOL ret;
-
-    if (!(ioctl = malloc( len ))) return FALSE;
-    ioctl->create_backup = FALSE;
-    ioctl->folder_offset = sizeof(*ioctl);
-    ioctl->folder_size = path->Length;
-    memcpy( (char *)ioctl + ioctl->folder_offset, path->Buffer, ioctl->folder_size );
-    ioctl->symlink_offset = ioctl->folder_offset + ioctl->folder_size;
-    strcpy( (char *)ioctl + ioctl->symlink_offset, link );
-
-    ret = DeviceIoControl( mgr, IOCTL_MOUNTMGR_DEFINE_SHELL_FOLDER, ioctl, len, NULL, 0, NULL, NULL );
-    free( ioctl );
-    return ret;
-}
-
-/******************************************************************************
- * create_link
- *
- * Sets up a symbolic link for one of the 'My Whatever' shell folders to point
- * into the corresponding XDG directory.
- */
-static void create_link( const WCHAR *path, const char *xdg_name, const char *default_name )
-{
-    UNICODE_STRING nt_name;
-    char *target = NULL;
-    HANDLE mgr;
-
-    if ((mgr = CreateFileW( MOUNTMGR_DOS_DEVICE_NAME, GENERIC_READ | GENERIC_WRITE,
-                            FILE_SHARE_READ | FILE_SHARE_WRITE, NULL, OPEN_EXISTING,
-                            0, 0 )) == INVALID_HANDLE_VALUE)
-    {
-        FIXME( "failed to connect to mount manager\n" );
-        return;
-    }
-
-    nt_name.Buffer = NULL;
-    if (!RtlDosPathNameToNtPathName_U( path, &nt_name, NULL, NULL )) goto done;
-
-    if ((target = get_xdg_path( xdg_name )))
-    {
-        if (link_folder( mgr, &nt_name, target )) goto done;
-    }
-    link_folder( mgr, &nt_name, default_name );
-
-done:
-    RtlFreeUnicodeString( &nt_name );
-    free( target );
-    CloseHandle( mgr );
-}
-
-/******************************************************************************
- * _SHCreateSymbolicLink  [Internal]
- *
- * Sets up a symbolic link for one of the special shell folders to point into
- * the users home directory.
- *
- * PARAMS
- *  nFolder [I] CSIDL identifying the folder.
- */
-static void _SHCreateSymbolicLink(int nFolder, const WCHAR *path)
-{
-    DWORD folder = nFolder & CSIDL_FOLDER_MASK;
-
-    switch (folder) {
-        case CSIDL_PERSONAL:
-            create_link( path, "XDG_DOCUMENTS_DIR", "$HOME/Documents" );
-            break;
-        case CSIDL_DESKTOPDIRECTORY:
-            create_link( path, "XDG_DESKTOP_DIR", "$HOME/Desktop" );
-            break;
-        case CSIDL_MYPICTURES:
-            create_link( path, "XDG_PICTURES_DIR", "$HOME/Pictures" );
-            break;
-        case CSIDL_MYVIDEO:
-            create_link( path, "XDG_VIDEOS_DIR", "$HOME/Movies" );
-            break;
-        case CSIDL_MYMUSIC:
-            create_link( path, "XDG_MUSIC_DIR", "$HOME/Music" );
-            break;
-        case CSIDL_DOWNLOADS:
-            create_link( path, "XDG_DOWNLOAD_DIR", "$HOME/Downloads" );
-            break;
-        case CSIDL_TEMPLATES:
-            create_link( path, "XDG_TEMPLATES_DIR", "$HOME/Templates" );
-            break;
-    }
 }
 
 /******************************************************************************
@@ -3027,11 +2858,6 @@
         goto end;
     }
 
-    /* create symbolic links rather than directories for specific
-     * user shell folders */
-    if (!pszSubPath)
-        _SHCreateSymbolicLink(folder, szBuildPath);
-
     /* create directory/directories */
     ret = SHCreateDirectoryExW(hwndOwner, szBuildPath, NULL);
     if (ret && ret != ERROR_ALREADY_EXISTS)
@@ -3086,6 +2912,8 @@
  LPCWSTR szUserShellFolderPath, LPCWSTR szShellFolderPath, const UINT folders[],
  UINT foldersLen)
 {
+    static const WCHAR WineVistaPathsW[] = {'_','_','W','i','n','e','V','i','s','t','a','P','a','t','h','s',0};
+
     const WCHAR *szValueName;
     WCHAR buffer[40];
     UINT i;
@@ -3094,6 +2922,7 @@
     HKEY hUserKey = NULL, hKey = NULL;
     DWORD dwType, dwPathLen;
     LONG ret;
+    DWORD already_vista_paths = 0;
 
     TRACE("%p,%p,%s,%p,%u\n", hRootKey, hToken,
      debugstr_w(szUserShellFolderPath), folders, foldersLen);
@@ -3107,6 +2936,12 @@
         if (ret)
             hr = HRESULT_FROM_WIN32(ret);
     }
+
+    /* check if the registry has already been updated to the vista+ style paths */
+    dwPathLen = sizeof(already_vista_paths);
+    RegQueryValueExW(hUserKey, WineVistaPathsW, NULL, &dwType,
+            (LPBYTE)&already_vista_paths, &dwPathLen);
+
     for (i = 0; SUCCEEDED(hr) && i < foldersLen; i++)
     {
         dwPathLen = MAX_PATH * sizeof(WCHAR);
@@ -3119,9 +2954,10 @@
             szValueName = &buffer[0];
         }
 
-        if (RegQueryValueExW(hUserKey, szValueName, NULL,
-         &dwType, (LPBYTE)path, &dwPathLen) || (dwType != REG_SZ &&
-         dwType != REG_EXPAND_SZ))
+        if (!already_vista_paths ||
+                RegQueryValueExW(hUserKey, szValueName, NULL, &dwType,
+                    (LPBYTE)path, &dwPathLen) ||
+                (dwType != REG_SZ && dwType != REG_EXPAND_SZ))
         {
             *path = '\0';
             if (CSIDL_Data[folders[i]].type == CSIDL_Type_User)
@@ -3162,6 +2998,11 @@
              hToken, SHGFP_TYPE_DEFAULT, path);
         }
     }
+
+    already_vista_paths = 1;
+    RegSetValueExW(hUserKey, WineVistaPathsW, 0, REG_DWORD,
+            (LPBYTE)&already_vista_paths, sizeof(already_vista_paths));
+
     if (hUserKey)
         RegCloseKey(hUserKey);
     if (hKey)
@@ -3311,6 +3152,23 @@
         hr = SHGetFolderPathAndSubDirW(0, CSIDL_MYPICTURES | CSIDL_FLAG_CREATE, NULL,
                                        SHGFP_TYPE_DEFAULT, L"Screenshots", path);
     }
+
+
+    /* Proton HACK: In older Proton versions, duplicate Stuff directories were
+     * created at both %PROFILE%\Music and %PROFILE\Documents\Music. Due to
+     * some bugs when downgrading to those older Proton versions, create those
+     * missing Documents directories here, too. */
+    SHGetFolderPathAndSubDirW(0, CSIDL_PERSONAL | CSIDL_FLAG_CREATE, NULL,
+                                   SHGFP_TYPE_DEFAULT, L"Downloads", path);
+    SHGetFolderPathAndSubDirW(0, CSIDL_PERSONAL | CSIDL_FLAG_CREATE, NULL,
+                                   SHGFP_TYPE_DEFAULT, L"Music", path);
+    SHGetFolderPathAndSubDirW(0, CSIDL_PERSONAL | CSIDL_FLAG_CREATE, NULL,
+                                   SHGFP_TYPE_DEFAULT, L"Pictures", path);
+    SHGetFolderPathAndSubDirW(0, CSIDL_PERSONAL | CSIDL_FLAG_CREATE, NULL,
+                                   SHGFP_TYPE_DEFAULT, L"Templates", path);
+    SHGetFolderPathAndSubDirW(0, CSIDL_PERSONAL | CSIDL_FLAG_CREATE, NULL,
+                                   SHGFP_TYPE_DEFAULT, L"Videos", path);
+
     return hr;
 }
 
@@ -3906,7 +3764,7 @@
     /* check is it registry-registered folder */
     hr = get_known_folder_registry_path(kfid, NULL, &knownfolder->registryPath);
     if(SUCCEEDED(hr))
-        hr = HRESULT_FROM_WIN32(RegOpenKeyExW(HKEY_LOCAL_MACHINE, knownfolder->registryPath, 0, KEY_ENUMERATE_SUB_KEYS, &hKey));
+        hr = HRESULT_FROM_WIN32(RegOpenKeyExW(HKEY_LOCAL_MACHINE, knownfolder->registryPath, 0, 0, &hKey));
 
     if(SUCCEEDED(hr))
     {
@@ -4352,7 +4210,7 @@
     hr = get_known_folder_registry_path(id, NULL, &registryPath);
     if(SUCCEEDED(hr))
     {
-        hr = HRESULT_FROM_WIN32(RegOpenKeyExW(HKEY_LOCAL_MACHINE, registryPath, 0, KEY_ENUMERATE_SUB_KEYS, &hKey));
+        hr = HRESULT_FROM_WIN32(RegOpenKeyExW(HKEY_LOCAL_MACHINE, registryPath, 0, 0, &hKey));
         free(registryPath);
     }
 
