diff --git a/dlls/user32/input.c b/dlls/user32/input.c
--- a/dlls/user32/input.c
+++ b/dlls/user32/input.c
@@ -31,6 +31,7 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(win);
 WINE_DECLARE_DEBUG_CHANNEL(keyboard);
+WINE_DECLARE_DEBUG_CHANNEL(rawinput);
 
 /***********************************************************************
  *           get_locale_kbd_layout
@@ -108,6 +109,24 @@
 HWND WINAPI GetCapture(void)
 {
     return (HWND)NtUserGetThreadState( UserThreadStateCaptureWindow );
+}
+
+
+/*****************************************************************
+ *		SetCaretPos (USER32.@)
+ */
+BOOL WINAPI SetCaretPos( int x, int y )
+{
+    return NtUserSetCaretPos( x, y );
+}
+
+
+/*****************************************************************
+ *		SetCaretBlinkTime (USER32.@)
+ */
+BOOL WINAPI SetCaretBlinkTime( unsigned int time )
+{
+    return NtUserSetCaretBlinkTime( time );
 }
 
 
@@ -453,6 +472,7 @@
 
 static DWORD CALLBACK devnotify_window_callbackW(HANDLE handle, DWORD flags, DEV_BROADCAST_HDR *header)
 {
+    TRACE_(rawinput)("handle %p, flags %#lx, header %p\n", handle, flags, header);
     SendMessageTimeoutW(handle, WM_DEVICECHANGE, flags, (LPARAM)header, SMTO_ABORTIFHUNG, 2000, NULL);
     return 0;
 }
@@ -515,6 +535,91 @@
     return 0;
 }
 
+static BOOL steam_input_get_vid_pid( UINT slot, UINT16 *vid, UINT16 *pid )
+{
+    const char *info = getenv( "SteamVirtualGamepadInfo" );
+    char buffer[256];
+    UINT current;
+    FILE *file;
+
+    TRACE( "reading SteamVirtualGamepadInfo %s\n", debugstr_a(info) );
+
+    if (!info || !(file = fopen( info, "r" ))) return FALSE;
+    while (fscanf( file, "%255[^\n]\n", buffer ) == 1)
+    {
+        if (sscanf( buffer, "[slot %d]", &current )) continue;
+        if (current < slot) continue;
+        if (current > slot) break;
+        if (sscanf( buffer, "VID=0x%hx", vid )) continue;
+        if (sscanf( buffer, "PID=0x%hx", pid )) continue;
+    }
+
+    fclose( file );
+
+    return TRUE;
+}
+
+/* CW-Bug-Id: #23185 Emulate Steam Input native hooks for native SDL */
+static BOOL steam_input_devnotify(HANDLE handle, DWORD flags, DEV_BROADCAST_HDR *header, BOOL ansi)
+{
+    char buffer[offsetof(DEV_BROADCAST_DEVICEINTERFACE_W, dbcc_name[MAX_PATH])];
+    DEV_BROADCAST_DEVICEINTERFACE_W *copyW = (DEV_BROADCAST_DEVICEINTERFACE_W *)buffer;
+
+    if (flags & 0x8000)
+    {
+        switch (header->dbch_devicetype)
+        {
+        case DBT_DEVTYP_DEVICEINTERFACE:
+        {
+            static const WCHAR steam_input_idW[] = L"\\\\?\\HID#VID_28DE&PID_11FF&IG_";
+            const DEV_BROADCAST_DEVICEINTERFACE_W *ifaceW = (const DEV_BROADCAST_DEVICEINTERFACE_W *)header;
+
+            if (!wcsnicmp( ifaceW->dbcc_name, steam_input_idW, 29 ))
+            {
+                UINT size, slot;
+                const WCHAR *tmpW;
+                UINT16 vid, pid;
+
+                copyW->dbcc_devicetype = ifaceW->dbcc_devicetype;
+                copyW->dbcc_reserved = ifaceW->dbcc_reserved;
+                copyW->dbcc_classguid = ifaceW->dbcc_classguid;
+
+                if (swscanf( ifaceW->dbcc_name + 29, L"%02u", &slot ) != 1) slot = 0;
+                if (!steam_input_get_vid_pid( slot, &vid, &pid ))
+                {
+                    vid = 0x045e;
+                    pid = 0x028e;
+                }
+
+                size = swprintf( copyW->dbcc_name, MAX_PATH, L"\\\\.\\pipe\\HID#VID_045E&PID_028E&IG_00#%04X&%04X", vid, pid );
+                if ((tmpW = wcschr( ifaceW->dbcc_name + 29, '&' )))
+                {
+                    do copyW->dbcc_name[size++] = *tmpW++;
+                    while (*tmpW != '&' && size < MAX_PATH);
+                }
+                size += swprintf( copyW->dbcc_name + size, MAX_PATH - size, L"#%d#%u", slot, (UINT)GetCurrentProcessId() );
+
+                copyW->dbcc_size = offsetof(DEV_BROADCAST_DEVICEINTERFACE_W, dbcc_name[size + 1]);
+                header = (DEV_BROADCAST_HDR *)copyW;
+            }
+        }
+        }
+    }
+
+    if (ansi) return devnotify_window_callbackA(handle, flags, header);
+    return devnotify_window_callbackW(handle, flags, header);
+}
+
+static DWORD CALLBACK steam_input_callbackW(HANDLE handle, DWORD flags, DEV_BROADCAST_HDR *header)
+{
+    return steam_input_devnotify(handle, flags, header, FALSE);
+}
+
+static DWORD CALLBACK steam_input_callbackA(HANDLE handle, DWORD flags, DEV_BROADCAST_HDR *header)
+{
+    return steam_input_devnotify(handle, flags, header, TRUE);
+}
+
 static DWORD CALLBACK devnotify_service_callback(HANDLE handle, DWORD flags, DEV_BROADCAST_HDR *header)
 {
     FIXME("Support for service handles is not yet implemented!\n");
@@ -539,7 +644,7 @@
     DEV_BROADCAST_HDR *header = filter;
     device_notify_callback callback;
 
-    TRACE("handle %p, filter %p, flags %#lx\n", handle, filter, flags);
+    TRACE_(rawinput)("handle %p, filter %p, flags %#lx\n", handle, filter, flags);
 
     if (flags & ~(DEVICE_NOTIFY_SERVICE_HANDLE | DEVICE_NOTIFY_ALL_INTERFACE_CLASSES))
     {
@@ -556,9 +661,9 @@
     if (flags & DEVICE_NOTIFY_SERVICE_HANDLE)
         callback = devnotify_service_callback;
     else if (IsWindowUnicode( handle ))
-        callback = devnotify_window_callbackW;
+        callback = steam_input_callbackW;
     else
-        callback = devnotify_window_callbackA;
+        callback = steam_input_callbackA;
 
     if (!header)
     {
@@ -577,7 +682,10 @@
         return I_ScRegisterDeviceNotification( handle, (DEV_BROADCAST_HDR *)&iface, callback );
     }
     if (header->dbch_devicetype == DBT_DEVTYP_HANDLE)
+    {
+        FIXME( "DBT_DEVTYP_HANDLE not implemented\n" );
         return I_ScRegisterDeviceNotification( handle, header, callback );
+    }
 
     FIXME( "type %#lx not implemented\n", header->dbch_devicetype );
     SetLastError( ERROR_INVALID_DATA );
@@ -646,28 +754,8 @@
  */
 BOOL WINAPI CloseTouchInputHandle( HTOUCHINPUT handle )
 {
-    FIXME( "handle %p stub!\n", handle );
-    SetLastError( ERROR_CALL_NOT_IMPLEMENTED );
-    return FALSE;
-}
-
-/*****************************************************************************
- * GetTouchInputInfo (USER32.@)
- */
-BOOL WINAPI GetTouchInputInfo( HTOUCHINPUT handle, UINT count, TOUCHINPUT *ptr, int size )
-{
-    FIXME( "handle %p, count %u, ptr %p, size %u stub!\n", handle, count, ptr, size );
-    SetLastError( ERROR_CALL_NOT_IMPLEMENTED );
-    return FALSE;
-}
-
-/**********************************************************************
- * IsTouchWindow (USER32.@)
- */
-BOOL WINAPI IsTouchWindow( HWND hwnd, ULONG *flags )
-{
-    FIXME( "hwnd %p, flags %p stub!\n", hwnd, flags );
-    return FALSE;
+    TRACE( "handle %p.\n", handle );
+    return TRUE;
 }
 
 /*****************************************************************************
@@ -675,8 +763,8 @@
  */
 BOOL WINAPI RegisterTouchWindow( HWND hwnd, ULONG flags )
 {
-    FIXME( "hwnd %p, flags %#lx stub!\n", hwnd, flags );
-    return TRUE;
+    TRACE( "hwnd %p, flags %#lx.\n", hwnd, flags );
+    return NtUserCallTwoParam( (ULONG_PTR)hwnd, flags, NtUserCallTwoParam_RegisterTouchWindow );
 }
 
 /*****************************************************************************
@@ -684,8 +772,8 @@
  */
 BOOL WINAPI UnregisterTouchWindow( HWND hwnd )
 {
-    FIXME( "hwnd %p stub!\n", hwnd );
-    return TRUE;
+    TRACE( "hwnd %p.\n", hwnd );
+    return NtUserCallOneParam( (ULONG_PTR)hwnd, NtUserCallOneParam_UnregisterTouchWindow );
 }
 
 /*****************************************************************************
@@ -742,45 +830,13 @@
     return FALSE;
 }
 
-BOOL WINAPI GetPointerDeviceProperties( HANDLE device, UINT32 *count,
-                                        POINTER_DEVICE_PROPERTY *properties)
-{
-    FIXME( "device %p, count %p, info %p stub!\n", device, count, properties );
-    SetLastError( ERROR_CALL_NOT_IMPLEMENTED );
-    return FALSE;
-}
-
-BOOL WINAPI GetPointerDeviceRects( HANDLE device, RECT *device_rect, RECT *display_rect )
-{
-    FIXME( "device %p, device_rect %p, display_rect %p stub!\n",
-           device, device_rect, display_rect );
-    SetLastError( ERROR_CALL_NOT_IMPLEMENTED );
-    return FALSE;
-}
-
-BOOL WINAPI GetPointerPenInfo( UINT32 id, POINTER_PEN_INFO *info )
+BOOL WINAPI GetPointerTouchInfo( UINT32 id, POINTER_TOUCH_INFO *info )
 {
     FIXME( "id %u, info %p stub!\n", id, info );
     SetLastError( ERROR_CALL_NOT_IMPLEMENTED );
     return FALSE;
 }
 
-BOOL WINAPI GetPointerTouchInfo( UINT32 id, POINTER_TOUCH_INFO *info )
-{
-    FIXME( "id %u, info %p stub!\n", id, info );
-    SetLastError( ERROR_CALL_NOT_IMPLEMENTED );
-    return FALSE;
-}
-
-BOOL WINAPI GetRawPointerDeviceData( UINT32 id, UINT32 hist_count, UINT32 prop_count,
-                                     POINTER_DEVICE_PROPERTY *properties, LONG *values )
-{
-    FIXME( "id %u, count %u, prop_count %u, properties %p, values %p stub!\n",
-           id, hist_count, prop_count, properties, values );
-    SetLastError( ERROR_CALL_NOT_IMPLEMENTED );
-    return FALSE;
-}
-
 BOOL WINAPI GetPointerTouchInfoHistory( UINT32 id, UINT32 *count, POINTER_TOUCH_INFO *info )
 {
     FIXME( "id %u, count %p, info %p stub!\n", id, count, info );
@@ -790,6 +846,15 @@
 
 
 /*******************************************************************
+ *           SetForegroundWindow  (USER32.@)
+ */
+BOOL WINAPI SetForegroundWindow( HWND hwnd )
+{
+    return NtUserSetForegroundWindow( hwnd );
+}
+
+
+/*******************************************************************
  *           GetActiveWindow  (USER32.@)
  */
 HWND WINAPI GetActiveWindow(void)
@@ -826,6 +891,15 @@
 
 
 /***********************************************************************
+ *           SetProgmanWindow (USER32.@)
+ */
+HWND WINAPI SetProgmanWindow( HWND hwnd )
+{
+    return NtUserSetProgmanWindow( hwnd );
+}
+
+
+/***********************************************************************
  *           GetProgmanWindow (USER32.@)
  */
 HWND WINAPI GetProgmanWindow(void)
@@ -833,6 +907,14 @@
     return NtUserGetProgmanWindow();
 }
 
+
+/***********************************************************************
+ *           SetTaskmanWindow (USER32.@)
+ */
+HWND WINAPI SetTaskmanWindow( HWND hwnd )
+{
+    return NtUserSetTaskmanWindow( hwnd );
+}
 
 /***********************************************************************
  *           GetTaskmanWindow (USER32.@)
