diff --git a/dlls/dbghelp/msc.c b/dlls/dbghelp/msc.c
--- a/dlls/dbghelp/msc.c
+++ b/dlls/dbghelp/msc.c
@@ -47,6 +47,8 @@
 #include "wine/mscvpdb.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(dbghelp_msc);
+
+static const GUID null_guid;
 
 struct pdb_stream_name
 {
@@ -2714,11 +2716,10 @@
                 }
                 else
                 {
-                    const union codeview_symbol* end_sym;
                     /* skip all records until paired S_INLINESITE_END */
-                    end_sym = (const union codeview_symbol*)(root + sym->inline_site_v3.pEnd);
-                    if (end_sym->generic.id != S_INLINESITE_END) FIXME("complete wreckage\n");
-                    length = sym->inline_site_v3.pEnd - i + end_sym->generic.len + 2;
+                    sym = (const union codeview_symbol*)(root + sym->inline_site_v3.pEnd);
+                    if (sym->generic.id != S_INLINESITE_END) FIXME("complete wreckage\n");
+                    length = sym->inline_site_v3.pEnd - i + sym->generic.len;
                 }
             }
             break;
@@ -2737,11 +2738,10 @@
                 }
                 else
                 {
-                    const union codeview_symbol* end_sym;
                     /* skip all records until paired S_INLINESITE_END */
-                    end_sym = (const union codeview_symbol*)(root + sym->inline_site2_v3.pEnd);
-                    if (end_sym->generic.id != S_INLINESITE_END) FIXME("complete wreckage\n");
-                    length = sym->inline_site2_v3.pEnd - i + end_sym->generic.len + 2;
+                    sym = (const union codeview_symbol*)(root + sym->inline_site2_v3.pEnd);
+                    if (sym->generic.id != S_INLINESITE_END) FIXME("complete wreckage\n");
+                    length = sym->inline_site2_v3.pEnd - i + sym->generic.len;
                 }
             }
             break;
@@ -3604,7 +3604,8 @@
     return pdb_file->stream_dict != NULL;
 }
 
-static BOOL pdb_process_internal(const struct msc_debug_info *msc_dbg,
+static BOOL pdb_process_internal(const struct process *pcs,
+                                 const struct msc_debug_info *msc_dbg,
                                  const WCHAR *filename,
                                  struct old_pdb_module_info *pdb_module_info,
                                  unsigned module_index,
@@ -3719,7 +3720,8 @@
     return ERROR_BAD_FORMAT;
 }
 
-static void pdb_process_symbol_imports(const struct msc_debug_info *msc_dbg,
+static void pdb_process_symbol_imports(const struct process *pcs,
+                                       const struct msc_debug_info *msc_dbg,
                                        const PDB_SYMBOLS *symbols,
                                        const void *symbols_image,
                                        const char *image,
@@ -3741,7 +3743,6 @@
         last = (const char*)imp + symbols->pdbimport_size;
         while (imp < (const PDB_SYMBOL_IMPORT*)last)
         {
-            WCHAR               buffer[MAX_PATH];
             SYMSRV_INDEX_INFOW info;
             BOOL line_info;
 
@@ -3749,10 +3750,9 @@
             if (i >= CV_MAX_MODULES) FIXME("Out of bounds!!!\n");
             TRACE("got for %s: age=%u ts=%x\n",
                   debugstr_a(imp->filename), imp->Age, imp->TimeDateStamp);
-            MultiByteToWideChar(CP_ACP, 0, imp->filename, -1, buffer, ARRAY_SIZE(buffer));
-            if (path_find_symbol_file(msc_dbg->module, buffer, TRUE, NULL, imp->TimeDateStamp, imp->Age, &info,
+            if (path_find_symbol_file(pcs, msc_dbg->module, imp->filename, TRUE, NULL, imp->TimeDateStamp, imp->Age, &info,
                                       &msc_dbg->module->module.PdbUnmatched))
-                pdb_process_internal(msc_dbg, info.pdbfile, pdb_module_info, i, &line_info);
+                pdb_process_internal(pcs, msc_dbg, info.pdbfile, pdb_module_info, i, &line_info);
             i++;
             imp = (const PDB_SYMBOL_IMPORT*)((const char*)first + ((ptr - (const char*)first + strlen(ptr) + 1 + 3) & ~3));
         }
@@ -3768,7 +3768,8 @@
     cv_current_module->allowed = TRUE;
 }
 
-static BOOL pdb_process_internal(const struct msc_debug_info *msc_dbg,
+static BOOL pdb_process_internal(const struct process *pcs,
+                                 const struct msc_debug_info *msc_dbg,
                                  const WCHAR *filename,
                                  struct old_pdb_module_info *pdb_module_info,
                                  unsigned module_index,
@@ -3843,7 +3844,7 @@
 
         files_image = pdb_read_strings(pdb_file);
 
-        pdb_process_symbol_imports(msc_dbg, &symbols, symbols_image, image,
+        pdb_process_symbol_imports(pcs, msc_dbg, &symbols, symbols_image, image,
                                    pdb_module_info, module_index);
         pdb_process_types(msc_dbg, pdb_file);
 
@@ -3941,7 +3942,7 @@
         pdb_free(ipi_image);
     }
     else
-        pdb_process_symbol_imports(msc_dbg, NULL, NULL, image,
+        pdb_process_symbol_imports(pcs, msc_dbg, NULL, NULL, image,
                                    pdb_module_info, module_index);
 
     pdb_free(symbols_image);
@@ -3964,7 +3965,8 @@
     pdb_location_compute,
 };
 
-static BOOL old_pdb_process_file(const struct msc_debug_info *msc_dbg,
+static BOOL old_pdb_process_file(const struct process *pcs,
+                                 const struct msc_debug_info *msc_dbg,
                                  const WCHAR *filename, BOOL *has_linenumber_info)
 {
     struct module_format*       modfmt;
@@ -3983,7 +3985,7 @@
 
     memset(cv_zmodules, 0, sizeof(cv_zmodules));
     codeview_init_basic_types(msc_dbg->module);
-    ret = pdb_process_internal(msc_dbg, filename,
+    ret = pdb_process_internal(pcs, msc_dbg, filename,
                                msc_dbg->module->format_info[DFI_OLD_PDB]->u.old_pdb_info, -1, has_linenumber_info);
     codeview_clear_type_table();
     if (!ret)
@@ -3993,6 +3995,54 @@
     }
     return ret;
 }
+
+static BOOL pdb_process_file(const struct process *pcs,
+                             const struct msc_debug_info *msc_dbg,
+                             const char *filename, const GUID *guid, DWORD timestamp, DWORD age)
+{
+    SYMSRV_INDEX_INFOW          info;
+    BOOL                        unmatched, has_linenumber_info, ret;
+
+    if (!msc_dbg->module->dont_load_symbols &&
+        path_find_symbol_file(pcs, msc_dbg->module, filename, TRUE, guid, timestamp, age, &info, &unmatched))
+    {
+        if (getenv("WINE_DBGHELP_OLD_PDB")) /* keep using old pdb reader */
+            ret = old_pdb_process_file(pcs, msc_dbg, info.pdbfile, &has_linenumber_info);
+        else
+            ret = pdb_init_modfmt(pcs, msc_dbg, info.pdbfile, &has_linenumber_info);
+
+        if (ret)
+        {
+            msc_dbg->module->module.SymType = SymPdb;
+            msc_dbg->module->module.PdbSig = info.sig;
+            msc_dbg->module->module.PdbAge = info.age;
+            msc_dbg->module->module.PdbSig70 = info.guid;
+            msc_dbg->module->module.PdbUnmatched = unmatched;
+            wcscpy(msc_dbg->module->module.LoadedPdbName, info.pdbfile);
+
+            /* FIXME: we could have a finer grain here */
+            msc_dbg->module->module.LineNumbers = has_linenumber_info;
+            msc_dbg->module->module.GlobalSymbols = TRUE;
+            msc_dbg->module->module.TypeInfo = TRUE;
+            msc_dbg->module->module.SourceIndexed = TRUE;
+            msc_dbg->module->module.Publics = TRUE;
+
+            return TRUE;
+        }
+    }
+    msc_dbg->module->module.SymType = SymNone;
+    if (guid)
+        msc_dbg->module->module.PdbSig70 = *guid;
+    else
+        memset(&msc_dbg->module->module.PdbSig70, 0, sizeof(GUID));
+    msc_dbg->module->module.PdbSig = 0;
+    msc_dbg->module->module.PdbAge = age;
+    return FALSE;
+}
+
+/*========================================================================
+ * Process CodeView debug information.
+ */
 
 #define MAKESIG(a,b,c,d)        ((a) | ((b) << 8) | ((c) << 16) | ((d) << 24))
 #define CODEVIEW_NB09_SIG       MAKESIG('N','B','0','9')
@@ -4000,43 +4050,8 @@
 #define CODEVIEW_NB11_SIG       MAKESIG('N','B','1','1')
 #define CODEVIEW_RSDS_SIG       MAKESIG('R','S','D','S')
 
-BOOL pdb_load_debug_info(struct module *module, const SYMSRV_INDEX_INFOW *info, BOOL unmatched)
-{
-    BOOL                        has_linenumber_info, ret;
-    struct msc_debug_info       msc_dbg = {.module = module};
-
-    msc_dbg.nsect = pe_clone_sections_table(module, (IMAGE_SECTION_HEADER**)&msc_dbg.sectp);
-    if (!msc_dbg.nsect) return FALSE;
-    if (getenv("WINE_DBGHELP_OLD_PDB")) /* keep using old pdb reader */
-        ret = old_pdb_process_file(&msc_dbg, info->pdbfile, &has_linenumber_info);
-    else
-        ret = pdb_init_modfmt(&msc_dbg, info->pdbfile, &has_linenumber_info);
-
-    HeapFree(GetProcessHeap(), 0, (void*)msc_dbg.sectp);
-    if (ret)
-    {
-        module->module.SymType = SymPdb;
-        module->module.PdbSig70 = info->guid;
-        module->module.PdbSig = info->sig;
-        module->module.PdbAge = info->age;
-        module->module.PdbUnmatched = unmatched;
-        wcscpy(module->module.LoadedPdbName, info->pdbfile);
-        module->module.CVSig = info->sig ? CODEVIEW_NB10_SIG : CODEVIEW_RSDS_SIG;
-        /* FIXME: we could have a finer grain here */
-        module->module.LineNumbers = has_linenumber_info;
-        module->module.GlobalSymbols = TRUE;
-        module->module.TypeInfo = TRUE;
-        module->module.SourceIndexed = TRUE;
-        module->module.Publics = TRUE;
-    }
-    return ret;
-}
-
-/*========================================================================
- * Process CodeView debug information.
- */
-
-static BOOL codeview_process_info(const struct msc_debug_info *msc_dbg)
+static BOOL codeview_process_info(const struct process *pcs,
+                                  const struct msc_debug_info *msc_dbg)
 {
     const DWORD*                signature = (const DWORD*)msc_dbg->root;
     BOOL                        ret = FALSE;
@@ -4130,22 +4145,45 @@
     }
 
     case CODEVIEW_NB10_SIG:
+    {
+        const CODEVIEW_PDB_DATA* pdb = (const CODEVIEW_PDB_DATA*)msc_dbg->root;
+        ret = pdb_process_file(pcs, msc_dbg, pdb->name, NULL, pdb->timestamp, pdb->age);
+        break;
+    }
     case CODEVIEW_RSDS_SIG:
-        /* should be loaded directly through pe_load_debug_info() */
-        break;
+    {
+        const OMFSignatureRSDS* rsds = (const OMFSignatureRSDS*)msc_dbg->root;
+
+        TRACE("Got RSDS type of PDB file: guid=%s age=%08x name=%s\n",
+              wine_dbgstr_guid(&rsds->guid), rsds->age, debugstr_a(rsds->name));
+        /* gcc/mingw and clang can emit build-id information, but with an empty PDB filename.
+         * Don't search for the .pdb file in that case.
+         */
+        if (rsds->name[0])
+            ret = pdb_process_file(pcs, msc_dbg, rsds->name, &rsds->guid, 0, rsds->age);
+        else
+            ret = TRUE;
+        break;
+    }
     default:
         ERR("Unknown CODEVIEW signature %08lx in module %s\n",
             *signature, debugstr_w(msc_dbg->module->modulename));
         break;
     }
-
+    if (ret)
+    {
+        msc_dbg->module->module.CVSig = *signature;
+        if (*signature == CODEVIEW_RSDS_SIG)
+            memcpy(msc_dbg->module->module.CVData, msc_dbg->root,
+                   sizeof(msc_dbg->module->module.CVData));
+    }
     return ret;
 }
 
 /*========================================================================
  * Process debug directory.
  */
-BOOL pe_load_debug_directory(struct module* module,
+BOOL pe_load_debug_directory(const struct process* pcs, struct module* module, 
                              const BYTE* mapping,
                              const IMAGE_SECTION_HEADER* sectp, DWORD nsect,
                              const IMAGE_DEBUG_DIRECTORY* dbg, int nDbg)
@@ -4181,7 +4219,7 @@
             if (dbg[i].Type == IMAGE_DEBUG_TYPE_CODEVIEW)
             {
                 msc_dbg.root = mapping + dbg[i].PointerToRawData;
-                if ((ret = codeview_process_info(&msc_dbg))) goto done;
+                if ((ret = codeview_process_info(pcs, &msc_dbg))) goto done;
             }
         }
     
@@ -4232,6 +4270,25 @@
     }
     __ENDTRY
 
+    /* we haven't found yet any debug information, fallback to unmatched pdb */
+    if (!ret && module->module.SymType == SymDeferred)
+    {
+        SYMSRV_INDEX_INFOW info = {.sizeofstruct = sizeof(info)};
+        char buffer[MAX_PATH];
+        char *ext;
+        DWORD options;
+
+        WideCharToMultiByte(CP_ACP, 0, module->module.LoadedImageName, -1, buffer, ARRAY_SIZE(buffer), 0, NULL);
+        ext = strrchr(buffer, '.');
+        if (ext) strcpy(ext + 1, "pdb"); else strcat(buffer, ".pdb");
+        options = SymGetOptions();
+        SymSetOptions(options | SYMOPT_LOAD_ANYTHING);
+        ret = pdb_process_file(pcs, &msc_dbg, buffer, &null_guid, 0, 0);
+        SymSetOptions(options);
+        if (!ret && module->dont_load_symbols)
+            module->module.TimeDateStamp = 0;
+    }
+
     return ret;
 }
 
