diff --git a/dlls/dwrite/main.c b/dlls/dwrite/main.c
--- a/dlls/dwrite/main.c
+++ b/dlls/dwrite/main.c
@@ -595,7 +595,6 @@
 struct dwritefactory
 {
     IDWriteFactory7 IDWriteFactory7_iface;
-    IDWriteFontDownloadQueue IDWriteFontDownloadQueue_iface;
     LONG refcount;
 
     IDWriteFontCollection *system_collections[DWRITE_FONT_FAMILY_MODEL_WEIGHT_STRETCH_STYLE + 1];
@@ -603,13 +602,6 @@
     IDWriteGdiInterop1 *gdiinterop;
     IDWriteFontFallback1 *fallback;
 
-    struct
-    {
-        FILETIME timestamp;
-        struct dwrite_fontset_entry **entries;
-        unsigned int count;
-    } system_set;
-
     IDWriteFontFileLoader *localfontfileloader;
     struct list localfontfaces;
 
@@ -643,15 +635,6 @@
     free(fileloader);
 }
 
-static void factory_cleanup_fontset(struct dwritefactory *factory)
-{
-    unsigned int i;
-
-    for (i = 0; i < factory->system_set.count; ++i)
-        release_fontset_entry(factory->system_set.entries[i]);
-    memset(&factory->system_set, 0, sizeof(factory->system_set));
-}
-
 static void release_dwritefactory(struct dwritefactory *factory)
 {
     struct fileloader *fileloader, *fileloader2;
@@ -676,7 +659,6 @@
         if (factory->system_collections[i])
             IDWriteFontCollection_Release(factory->system_collections[i]);
     }
-    factory_cleanup_fontset(factory);
     if (factory->eudc_collection)
         IDWriteFontCollection1_Release(factory->eudc_collection);
     if (factory->fallback)
@@ -840,7 +822,7 @@
     if (FAILED(hr))
         return hr;
 
-    hr = create_font_collection(iface, enumerator, (IDWriteFontCollection3 **)collection);
+    hr = create_font_collection(iface, enumerator, FALSE, (IDWriteFontCollection3 **)collection);
     IDWriteFontFileEnumerator_Release(enumerator);
     return hr;
 }
@@ -1541,16 +1523,6 @@
     return hr;
 }
 
-static HKEY open_fonts_key(void)
-{
-    HKEY hkey = NULL;
-
-    RegOpenKeyExW(HKEY_LOCAL_MACHINE, L"Software\\Microsoft\\Windows NT\\CurrentVersion\\Fonts",
-            0, GENERIC_READ | KEY_QUERY_VALUE, &hkey);
-
-    return hkey;
-}
-
 static HRESULT create_system_path_list(WCHAR ***ret, unsigned int *ret_count)
 {
     unsigned int index = 0, value_size, max_name_count;
@@ -1560,8 +1532,11 @@
     HKEY hkey;
     LONG r;
 
-    if (!(hkey = open_fonts_key()))
+    if (RegOpenKeyExA(HKEY_LOCAL_MACHINE, "Software\\Microsoft\\Windows NT\\CurrentVersion\\Fonts",
+            0, GENERIC_READ, &hkey))
+    {
         return E_UNEXPECTED;
+    }
 
     value_size = MAX_PATH * sizeof(*value);
     value = malloc(value_size);
@@ -1640,37 +1615,23 @@
     return S_OK;
 }
 
-static void get_system_fonts_mtime(FILETIME *mtime)
-{
-    HKEY hkey;
-
-    memset(mtime, 0, sizeof(*mtime));
-
-    if ((hkey = open_fonts_key()))
-    {
-        RegQueryInfoKeyW(hkey, NULL, NULL, NULL, NULL, NULL, NULL,
-                NULL, NULL, NULL, NULL, mtime);
-        RegCloseKey(hkey);
-    }
-}
-
 static int __cdecl create_system_fontset_compare(const void *left, const void *right)
 {
     const WCHAR *_l = *(WCHAR **)left, *_r = *(WCHAR **)right;
     return wcsicmp(_l, _r);
 };
 
-static HRESULT factory_create_system_fontset(struct dwritefactory *factory, const FILETIME *timestamp)
+HRESULT create_system_fontset(IDWriteFactory7 *factory, REFIID riid, void **obj)
 {
     IDWriteFontSetBuilder2 *builder;
+    IDWriteFontSet *fontset;
     unsigned int i, j, count;
     WCHAR **paths;
     HRESULT hr;
 
-    factory_cleanup_fontset(factory);
-    factory->system_set.timestamp = *timestamp;
-
-    if (FAILED(hr = create_fontset_builder(&factory->IDWriteFactory7_iface, TRUE, &builder))) return hr;
+    *obj = NULL;
+
+    if (FAILED(hr = create_fontset_builder(factory, &builder))) return hr;
 
     if (SUCCEEDED(hr = create_system_path_list(&paths, &count)))
     {
@@ -1693,42 +1654,14 @@
         free(paths);
     }
 
-    hr = fontset_builder_get_entries(builder, &factory->system_set.entries, &factory->system_set.count);
-    IDWriteFontSetBuilder2_Release(builder);
-
-    return hr;
-}
-
-HRESULT create_system_fontset(IDWriteFactory7 *factory_iface, REFIID riid, void **obj)
-{
-    struct dwritefactory *factory = impl_from_IDWriteFactory7(factory_iface);
-    IDWriteFontSet *fontset;
-    FILETIME timestamp;
-    HRESULT hr = S_OK;
-
-    *obj = NULL;
-
-    get_system_fonts_mtime(&timestamp);
-
-    EnterCriticalSection(&factory->cs);
-
-    if (CompareFileTime(&timestamp, &factory->system_set.timestamp) > 0)
-        hr = factory_create_system_fontset(factory, &timestamp);
-
-    if (SUCCEEDED(hr))
-    {
-        hr = fontset_create_from_set(factory_iface, factory->system_set.entries,
-                factory->system_set.count, TRUE, &fontset);
-    }
-
-    LeaveCriticalSection(&factory->cs);
-
-    if (SUCCEEDED(hr))
+    if (SUCCEEDED(hr = IDWriteFontSetBuilder2_CreateFontSet(builder, &fontset)))
     {
         hr = IDWriteFontSet_QueryInterface(fontset, riid, obj);
         IDWriteFontSet_Release(fontset);
     }
 
+    IDWriteFontSetBuilder2_Release(builder);
+
     return hr;
 }
 
@@ -1743,7 +1676,7 @@
 {
     TRACE("%p, %p.\n", iface, builder);
 
-    return create_fontset_builder(iface, FALSE, (IDWriteFontSetBuilder2 **)builder);
+    return create_fontset_builder(iface, (IDWriteFontSetBuilder2 **)builder);
 }
 
 static HRESULT WINAPI dwritefactory3_CreateFontCollectionFromFontSet(IDWriteFactory7 *iface, IDWriteFontSet *fontset,
@@ -1774,13 +1707,9 @@
 
 static HRESULT WINAPI dwritefactory3_GetFontDownloadQueue(IDWriteFactory7 *iface, IDWriteFontDownloadQueue **queue)
 {
-    struct dwritefactory *factory = impl_from_IDWriteFactory7(iface);
-
-    TRACE("%p, %p.\n", iface, queue);
-
-    *queue = &factory->IDWriteFontDownloadQueue_iface;
-    IDWriteFontDownloadQueue_AddRef(*queue);
-    return S_OK;
+    FIXME("%p, %p: stub\n", iface, queue);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI dwritefactory4_TranslateColorGlyphRun(IDWriteFactory7 *iface, D2D1_POINT_2F origin,
@@ -1862,7 +1791,7 @@
 {
     TRACE("%p, %p.\n", iface, builder);
 
-    return create_fontset_builder(iface, FALSE, (IDWriteFontSetBuilder2 **)builder);
+    return create_fontset_builder(iface, (IDWriteFontSetBuilder2 **)builder);
 }
 
 static HRESULT WINAPI dwritefactory5_CreateInMemoryFontFileLoader(IDWriteFactory7 *iface,
@@ -1951,7 +1880,7 @@
 {
     TRACE("%p, %p.\n", iface, builder);
 
-    return create_fontset_builder(iface, FALSE, builder);
+    return create_fontset_builder(iface, builder);
 }
 
 static HRESULT WINAPI dwritefactory6_CreateTextFormat(IDWriteFactory7 *iface, const WCHAR *family_name,
@@ -2145,96 +2074,10 @@
     dwritefactory7_GetSystemFontCollection,
 };
 
-static inline struct dwritefactory *impl_from_IDWriteFontDownloadQueue(IDWriteFontDownloadQueue *iface)
-{
-    return CONTAINING_RECORD(iface, struct dwritefactory, IDWriteFontDownloadQueue_iface);
-}
-
-static HRESULT WINAPI dwritefontdownloadqueue_QueryInterface(IDWriteFontDownloadQueue *iface,
-                                                             REFIID iid, void **out)
-{
-    TRACE("iface %p, iid %s, out %p.\n", iface, debugstr_guid(iid), out);
-
-    if (IsEqualGUID(iid, &IID_IUnknown) || IsEqualGUID(iid, &IID_IDWriteFontDownloadQueue))
-    {
-        *out = iface;
-        IDWriteFontDownloadQueue_AddRef(iface);
-        return S_OK;
-    }
-
-    *out = NULL;
-    WARN("%s not implemented, returning E_NOINTERFACE.\n", debugstr_guid(iid));
-    return E_NOINTERFACE;
-}
-
-static ULONG WINAPI dwritefontdownloadqueue_AddRef(IDWriteFontDownloadQueue *iface)
-{
-    struct dwritefactory *factory = impl_from_IDWriteFontDownloadQueue(iface);
-    return IDWriteFactory7_AddRef(&factory->IDWriteFactory7_iface);
-}
-
-static ULONG WINAPI dwritefontdownloadqueue_Release(IDWriteFontDownloadQueue *iface)
-{
-    struct dwritefactory *factory = impl_from_IDWriteFontDownloadQueue(iface);
-    return IDWriteFactory7_Release(&factory->IDWriteFactory7_iface);
-}
-
-static HRESULT WINAPI dwritefontdownloadqueue_AddListener(IDWriteFontDownloadQueue *iface,
-        IDWriteFontDownloadListener *listener, UINT32 *token)
-{
-    FIXME("%p, %p, %p stub!\n", iface, listener, token);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI dwritefontdownloadqueue_RemoveListener(IDWriteFontDownloadQueue *iface, UINT32 token)
-{
-    FIXME("%p, %#x stub!\n", iface, token);
-    return E_NOTIMPL;
-}
-
-static BOOL WINAPI dwritefontdownloadqueue_IsEmpty(IDWriteFontDownloadQueue *iface)
-{
-    FIXME("%p stub!\n", iface);
-    return TRUE;
-}
-
-static HRESULT WINAPI dwritefontdownloadqueue_BeginDownload(IDWriteFontDownloadQueue *iface, IUnknown *context)
-{
-    FIXME("%p, %p stub!\n", iface, context);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI dwritefontdownloadqueue_CancelDownload(IDWriteFontDownloadQueue *iface)
-{
-    FIXME("%p stub!\n", iface);
-    return E_NOTIMPL;
-}
-
-static UINT64 WINAPI dwritefontdownloadqueue_GetGenerationCount(IDWriteFontDownloadQueue *iface)
-{
-    FIXME("%p stub!\n", iface);
-    return 0;
-}
-
-static const struct IDWriteFontDownloadQueueVtbl dwritefontdownloadqueue_vtbl =
-{
-    dwritefontdownloadqueue_QueryInterface,
-    dwritefontdownloadqueue_AddRef,
-    dwritefontdownloadqueue_Release,
-    /* IDWriteFontDownloadQueue methods */
-    dwritefontdownloadqueue_AddListener,
-    dwritefontdownloadqueue_RemoveListener,
-    dwritefontdownloadqueue_IsEmpty,
-    dwritefontdownloadqueue_BeginDownload,
-    dwritefontdownloadqueue_CancelDownload,
-    dwritefontdownloadqueue_GetGenerationCount,
-};
-
 static void init_dwritefactory(struct dwritefactory *factory, DWRITE_FACTORY_TYPE type)
 {
     factory->IDWriteFactory7_iface.lpVtbl = type == DWRITE_FACTORY_TYPE_SHARED ?
             &shareddwritefactoryvtbl : &dwritefactoryvtbl;
-    factory->IDWriteFontDownloadQueue_iface.lpVtbl = &dwritefontdownloadqueue_vtbl;
     factory->refcount = 1;
     factory->localfontfileloader = get_local_fontfile_loader();
 
