diff --git a/dlls/mfplat/tests/mfplat.c b/dlls/mfplat/tests/mfplat.c
--- a/dlls/mfplat/tests/mfplat.c
+++ b/dlls/mfplat/tests/mfplat.c
@@ -60,23 +60,6 @@
 #include "mfd3d12.h"
 #include "wmcodecdsp.h"
 #include "dvdmedia.h"
-
-static void run_child_test(const char *name)
-{
-    char path_name[MAX_PATH];
-    PROCESS_INFORMATION info;
-    STARTUPINFOA startup;
-    char **argv;
-
-    winetest_get_mainargs(&argv);
-
-    memset(&startup, 0, sizeof(startup));
-    startup.cb = sizeof(startup);
-    sprintf(path_name, "%s mfplat %s", argv[0], name);
-    ok(CreateProcessA( NULL, path_name, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &info),
-            "CreateProcess failed.\n" );
-    wait_child_process(&info);
-}
 
 DEFINE_GUID(DUMMY_CLSID, 0x12345678,0x1234,0x1234,0x12,0x13,0x14,0x15,0x16,0x17,0x18,0x19);
 DEFINE_GUID(DUMMY_GUID1, 0x12345678,0x1234,0x1234,0x21,0x21,0x21,0x21,0x21,0x21,0x21,0x21);
@@ -889,8 +872,8 @@
 {
     struct test_callback *callback = impl_from_IMFAsyncCallback(iface);
     IMFSourceResolver *resolver;
+    IUnknown *object, *object2;
     MF_OBJECT_TYPE obj_type;
-    IUnknown *object;
     HRESULT hr;
 
     ok(!!result, "Unexpected result object.\n");
@@ -901,8 +884,13 @@
     hr = IMFSourceResolver_EndCreateObjectFromURL(resolver, result, &obj_type, &object);
     ok(hr == S_OK, "Failed to create an object, hr %#lx.\n", hr);
 
+    hr = IMFAsyncResult_GetObject(result, &object2);
+    ok(hr == S_OK, "Failed to get result object, hr %#lx.\n", hr);
+    ok(object2 == object, "Unexpected object.\n");
+
     if (object)
         IUnknown_Release(object);
+    IUnknown_Release(object2);
 
     SetEvent(callback->event);
 
@@ -1041,9 +1029,6 @@
 
             break;
         }
-
-        IMFMediaEvent_Release(callback->media_event);
-        callback->media_event = NULL;
     }
 
     if (callback->media_event)
@@ -1228,7 +1213,6 @@
         IMFStreamDescriptor_Release(sd);
 
         IMFPresentationDescriptor_Release(descriptor);
-        IMFMediaSource_Shutdown(source);
         IMFMediaSource_Release(source);
         IMFByteStream_Release(stream);
 
@@ -1376,8 +1360,8 @@
     ok(mediasource != NULL, "got %p\n", mediasource);
     ok(obj_type == MF_OBJECT_MEDIASOURCE, "got %d\n", obj_type);
 
-    IMFMediaSource_Shutdown(mediasource);
     refcount = IMFMediaSource_Release(mediasource);
+    todo_wine
     ok(!refcount, "Unexpected refcount %ld\n", refcount);
     IMFByteStream_Release(stream);
 
@@ -1389,11 +1373,7 @@
     hr = IMFSourceResolver_CreateObjectFromByteStream(resolver, stream, NULL, MF_RESOLUTION_MEDIASOURCE, NULL,
             &obj_type, (IUnknown **)&mediasource);
     ok(hr == S_OK || broken(hr == MF_E_UNSUPPORTED_BYTESTREAM_TYPE) /* w7 || w8 */, "Unexpected hr %#lx.\n", hr);
-    if (hr == S_OK)
-    {
-        IMFMediaSource_Shutdown(mediasource);
-        IMFMediaSource_Release(mediasource);
-    }
+    if (hr == S_OK) IMFMediaSource_Release(mediasource);
     IMFByteStream_Release(stream);
 
     hr = MFCreateFile(MF_ACCESSMODE_READ, MF_OPENMODE_FAIL_IF_NOT_EXIST, MF_FILEFLAGS_NONE, filename, &stream);
@@ -1405,11 +1385,7 @@
     hr = IMFSourceResolver_CreateObjectFromByteStream(resolver, stream, NULL, MF_RESOLUTION_MEDIASOURCE, NULL,
             &obj_type, (IUnknown **)&mediasource);
     ok(hr == S_OK || broken(hr == MF_E_UNSUPPORTED_BYTESTREAM_TYPE) /* w7 || w8 */, "Unexpected hr %#lx.\n", hr);
-    if (hr == S_OK)
-    {
-        IMFMediaSource_Shutdown(mediasource);
-        IMFMediaSource_Release(mediasource);
-    }
+    if (hr == S_OK) IMFMediaSource_Release(mediasource);
     IMFByteStream_Release(stream);
 
     /* stream must have a valid header, media cannot start in the middle of a stream */
@@ -5665,7 +5641,6 @@
     { &MFVideoFormat_NV12, 3, 2, 12, 9, 192, 9, 64 },
     { &MFVideoFormat_NV12, 4, 2, 12, 0, 192, 12, 64 },
     { &MFVideoFormat_NV12, 320, 240, 115200, 0, 115200, 115200, 320 },
-    { &MFVideoFormat_NV12, 650, 850, 828750, 0, 897600, 828750, 704 },
 
     /* YUV 4:2:0, 12 bpp, planar, half stride (the secondary plane has
      * the same height, half the width and half the stride of the
@@ -5738,13 +5713,6 @@
     { &MFVideoFormat_IYUV, 4, 2, 12, 0, 384, 12, 128 },
     { &MFVideoFormat_IYUV, 4, 3, 18, 0, 576, 18, 128 },
     { &MFVideoFormat_IYUV, 320, 240, 115200, 0, 138240, 115200, 384 },
-
-    { &MFVideoFormat_P010, 1, 2, 12, 6, 192, 6, 64 },
-    { &MFVideoFormat_P010, 2, 2, 12, 0, 192, 12, 64 },
-    { &MFVideoFormat_P010, 2, 4, 24, 0, 384, 24, 64 },
-    { &MFVideoFormat_P010, 3, 2, 24, 18, 192, 18, 64 },
-    { &MFVideoFormat_P010, 4, 2, 24, 0, 192, 24, 64 },
-    { &MFVideoFormat_P010, 320, 240, 230400, 0, 230400, 230400, 640 },
 };
 
 static void test_MFCalculateImageSize(void)
@@ -6402,9 +6370,6 @@
         return;
     }
 
-    hr = MFStartup(MF_VERSION, MFSTARTUP_FULL);
-    ok(hr == S_OK, "Failed to start up, hr %#lx.\n", hr);
-
     hr = pMFRegisterLocalSchemeHandler(NULL, NULL);
     ok(hr == E_INVALIDARG, "Unexpected hr %#lx.\n", hr);
 
@@ -6434,9 +6399,6 @@
 
     hr = pMFRegisterLocalByteStreamHandler(localW, localW, &local_activate);
     ok(hr == S_OK, "Failed to register stream handler, hr %#lx.\n", hr);
-
-    hr = MFShutdown();
-    ok(hr == S_OK, "Failed to shut down, hr %#lx.\n", hr);
 }
 
 static void test_create_property_store(void)
@@ -7178,7 +7140,9 @@
         sprintf(path_name, "%s mfplat s%d", argv[0], system_queues[i]);
         ok(CreateProcessA( NULL, path_name, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &info),
                 "CreateProcess failed.\n" );
-        wait_child_process(&info);
+        wait_child_process(info.hProcess);
+        CloseHandle(info.hProcess);
+        CloseHandle(info.hThread);
     }
 
     for (i = 0; i < ARRAY_SIZE(user_queues); ++i)
@@ -7188,7 +7152,9 @@
         sprintf(path_name, "%s mfplat u%d", argv[0], user_queues[i]);
         ok(CreateProcessA( NULL, path_name, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &info),
                 "CreateProcess failed.\n" );
-        wait_child_process(&info);
+        wait_child_process(info.hProcess);
+        CloseHandle(info.hProcess);
+        CloseHandle(info.hThread);
     }
 }
 
@@ -8153,6 +8119,40 @@
     hr = IMF2DBuffer2_Unlock2D(buffer_2d2);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     IMF2DBuffer2_Release(buffer_2d2);
+
+    IMFMediaBuffer_Release(buffer);
+
+    hr = IMFMediaType_SetGUID(media_type, &MF_MT_SUBTYPE, &MFVideoFormat_NV12);
+    ok(hr == S_OK, "Failed to set attribute, hr %#lx.\n", hr);
+    hr = IMFMediaType_SetUINT64(media_type, &MF_MT_FRAME_SIZE, (UINT64)96 << 32 | 96);
+    ok(hr == S_OK, "Failed to set attribute, hr %#lx.\n", hr);
+    hr = IMFMediaType_SetUINT32(media_type, &MF_MT_DEFAULT_STRIDE, 96);
+    ok(hr == S_OK, "Failed to set attribute, hr %#lx.\n", hr);
+    hr = pMFCreateMediaBufferFromMediaType(media_type, 0, 0, 0, &buffer);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+
+    hr = IMFMediaBuffer_QueryInterface(buffer, &IID_IMF2DBuffer, (void **)&buffer_2d);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    hr = IMF2DBuffer_Lock2D(buffer_2d, &data, &pitch);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    ok(pitch == 128, "got pitch %ld.\n", pitch);
+    hr = IMF2DBuffer_Unlock2D(buffer_2d);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    IMF2DBuffer_Release(buffer_2d);
+
+    IMFMediaBuffer_Release(buffer);
+
+    /* A linear buffer is created for YUV if MF_MT_DEFAULT_STRIDE is negative */
+    hr = IMFMediaType_SetUINT32(media_type, &MF_MT_DEFAULT_STRIDE, -128);
+    ok(hr == S_OK, "Failed to set attribute, hr %#lx.\n", hr);
+    hr = pMFCreateMediaBufferFromMediaType(media_type, 0, 0, 0, &buffer);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+
+    hr = IMFMediaBuffer_QueryInterface(buffer, &IID_IMF2DBuffer, (void **)&buffer_2d);
+    todo_wine
+    ok(hr == E_NOINTERFACE, "Unexpected hr %#lx.\n", hr);
+    if (SUCCEEDED(hr))
+        IMF2DBuffer_Release(buffer_2d);
 
     IMFMediaBuffer_Release(buffer);
 
@@ -13293,50 +13293,11 @@
         { &MEDIASUBTYPE_h264, &MEDIASUBTYPE_h264 },
         { &MEDIASUBTYPE_H264, &MFVideoFormat_H264 },
     };
-    static const GUID *audio_types[] =
-    {
-        &MEDIASUBTYPE_MP3,
-        &MEDIASUBTYPE_MSAUDIO1,
-        &MEDIASUBTYPE_WMAUDIO2,
-        &MEDIASUBTYPE_WMAUDIO3,
-        &MEDIASUBTYPE_WMAUDIO_LOSSLESS,
-        &MEDIASUBTYPE_PCM,
-        &MEDIASUBTYPE_IEEE_FLOAT,
-        &DUMMY_CLSID,
-    };
     MFVideoArea aperture;
     unsigned int i;
 
     hr = MFCreateMediaType(&media_type);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    memset(&mt, 0, sizeof(mt));
-    mt.majortype = MEDIATYPE_Audio;
-
-    for (i = 0; i < ARRAY_SIZE(audio_types); i++)
-    {
-        mt.subtype = *audio_types[i];
-
-        hr = MFInitMediaTypeFromAMMediaType(media_type, &mt);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-        hr = IMFMediaType_GetCount(media_type, &value32);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        ok(value32 == 4, "Unexpected value %#x.\n", value32);
-
-        hr = IMFMediaType_GetGUID(media_type, &MF_MT_MAJOR_TYPE, &guid);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        ok(IsEqualGUID(&guid, &MFMediaType_Audio), "Unexpected guid %s.\n", debugstr_guid(&guid));
-        hr = IMFMediaType_GetGUID(media_type, &MF_MT_SUBTYPE, &guid);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        ok(IsEqualGUID(&guid, audio_types[i]), "Unexpected guid %s.\n", debugstr_guid(&guid));
-        hr = IMFMediaType_GetUINT32(media_type, &MF_MT_ALL_SAMPLES_INDEPENDENT, &value32);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        ok(value32 == 1, "Unexpected value %#x.\n", value32);
-        hr = IMFMediaType_GetGUID(media_type, &MF_MT_AM_FORMAT_TYPE, &guid);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        ok(IsEqualGUID(&guid, &GUID_NULL), "Unexpected guid %s.\n", debugstr_guid(&guid));
-    }
 
     memset(&mt, 0, sizeof(mt));
     mt.majortype = MEDIATYPE_Video;
@@ -13828,20 +13789,19 @@
     ok(res == 0, "got %#lx\n", res);
     IMFAsyncResult_Release(result);
 
-    hr = MFPutWorkItem(0xffff, &callback->IMFAsyncCallback_iface, NULL);
+    hr = MFPutWorkItem(MFASYNC_CALLBACK_QUEUE_PRIVATE_MASK, &callback->IMFAsyncCallback_iface, NULL);
     ok(hr == S_OK, "got %#lx\n", hr);
     res = wait_async_callback_result(&callback->IMFAsyncCallback_iface, 100, &result);
     ok(res == 0, "got %#lx\n", res);
     IMFAsyncResult_Release(result);
 
-    hr = MFPutWorkItem(0x4000, &callback->IMFAsyncCallback_iface, NULL);
+    hr = MFPutWorkItem(MFASYNC_CALLBACK_QUEUE_PRIVATE_MASK & (MFASYNC_CALLBACK_QUEUE_PRIVATE_MASK - 1),
+            &callback->IMFAsyncCallback_iface, NULL);
     ok(hr == S_OK, "got %#lx\n", hr);
     res = wait_async_callback_result(&callback->IMFAsyncCallback_iface, 100, &result);
     ok(res == 0, "got %#lx\n", res);
     IMFAsyncResult_Release(result);
 
-    hr = MFPutWorkItem(0x10000, &callback->IMFAsyncCallback_iface, NULL);
-    ok(hr == MF_E_INVALID_WORKQUEUE, "got %#lx\n", hr);
     IMFAsyncCallback_Release(&callback->IMFAsyncCallback_iface);
 
     hr = MFShutdown();
@@ -13858,14 +13818,7 @@
     argc = winetest_get_mainargs(&argv);
     if (argc >= 3)
     {
-        if (!strcmp(argv[2], "startup"))
-            test_startup();
-        else if (!strcmp(argv[2], "startup_counts"))
-            test_startup_counts();
-        else if (!strcmp(argv[2], "undefined_queue_id"))
-            test_undefined_queue_id();
-        else
-            test_queue_com_state(argv[2]);
+        test_queue_com_state(argv[2]);
         return;
     }
 
@@ -13877,9 +13830,8 @@
 
     CoInitialize(NULL);
 
-    run_child_test("startup");
-    run_child_test("startup_counts");
-    run_child_test("undefined_queue_id");
+    test_startup();
+    test_startup_counts();
     test_register();
     test_media_type();
     test_MFCreateMediaEvent();
@@ -13951,6 +13903,7 @@
     test_MFInitMediaTypeFromAMMediaType();
     test_MFCreatePathFromURL();
     test_2dbuffer_copy();
+    test_undefined_queue_id();
     test_xvp_playback_mode();
 
     CoUninitialize();
