diff --git a/dlls/mfmediaengine/main.c b/dlls/mfmediaengine/main.c
--- a/dlls/mfmediaengine/main.c
+++ b/dlls/mfmediaengine/main.c
@@ -184,6 +184,7 @@
         BYTE *buffer;
         UINT buffer_size;
         DXGI_FORMAT output_format;
+        BOOL format_mismatch;
 
         struct
         {
@@ -814,79 +815,61 @@
     return a;
 }
 
-static void media_engine_get_frame_size(struct media_engine *engine, IMFTopology *topology)
+static void media_engine_get_frame_size(struct media_engine *engine)
 {
     IMFMediaTypeHandler *handler;
     IMFMediaType *media_type;
-    IMFStreamDescriptor *sd;
-    IMFTopologyNode *node;
-    unsigned int gcd;
-    UINT64 size;
-    HRESULT hr;
 
     engine->video_frame.size.cx = 0;
     engine->video_frame.size.cy = 0;
     engine->video_frame.ratio.cx = 1;
     engine->video_frame.ratio.cy = 1;
 
-    if (FAILED(IMFTopology_GetNodeByID(topology, engine->video_frame.node_id, &node)))
-        return;
-
-    hr = IMFTopologyNode_GetUnknown(node, &MF_TOPONODE_STREAM_DESCRIPTOR,
-            &IID_IMFStreamDescriptor, (void **)&sd);
-    IMFTopologyNode_Release(node);
-    if (FAILED(hr))
-        return;
-
-    hr = IMFStreamDescriptor_GetMediaTypeHandler(sd, &handler);
-    IMFStreamDescriptor_Release(sd);
-    if (FAILED(hr))
-        return;
-
-    hr = IMFMediaTypeHandler_GetCurrentMediaType(handler, &media_type);
-    IMFMediaTypeHandler_Release(handler);
-    if (FAILED(hr))
-    {
-        WARN("Failed to get current media type %#lx.\n", hr);
-        return;
-    }
-
-    IMFMediaType_GetUINT64(media_type, &MF_MT_FRAME_SIZE, &size);
-
-    engine->video_frame.size.cx = size >> 32;
-    engine->video_frame.size.cy = size;
-
-    if ((gcd = get_gcd(engine->video_frame.size.cx, engine->video_frame.size.cy)))
-    {
-        engine->video_frame.ratio.cx = engine->video_frame.size.cx / gcd;
-        engine->video_frame.ratio.cy = engine->video_frame.size.cy / gcd;
-    }
-
-    IMFMediaType_Release(media_type);
+    if (engine->presentation.frame_sink &&
+                SUCCEEDED(video_frame_sink_query_iface(engine->presentation.frame_sink, &IID_IMFMediaTypeHandler, (void**)&handler)))
+    {
+        if (SUCCEEDED(IMFMediaTypeHandler_GetCurrentMediaType(handler, &media_type)))
+        {
+            UINT64 size;
+            HRESULT hr = IMFMediaType_GetUINT64(media_type, &MF_MT_FRAME_SIZE, &size);
+            if (SUCCEEDED(hr))
+            {
+                unsigned int gcd;
+                engine->video_frame.size.cx = size >> 32;
+                engine->video_frame.size.cy = size;
+
+                if ((gcd = get_gcd(engine->video_frame.size.cx, engine->video_frame.size.cy)))
+                {
+                    engine->video_frame.ratio.cx = engine->video_frame.size.cx / gcd;
+                    engine->video_frame.ratio.cy = engine->video_frame.size.cy / gcd;
+                }
+            }
+            else
+            {
+                WARN("Failed to get frame size %#lx.\n", hr);
+            }
+
+            IMFMediaType_Release(media_type);
+        }
+        IMFMediaTypeHandler_Release(handler);
+    }
 }
 
 static void media_engine_apply_volume(const struct media_engine *engine)
 {
-    IMFAudioStreamVolume *as_volume;
-    UINT32 count = 0;
-    unsigned int i;
+    IMFSimpleAudioVolume *sa_volume;
     HRESULT hr;
 
     if (!engine->session)
         return;
 
-    if (FAILED(MFGetService((IUnknown *)engine->session, &MR_STREAM_VOLUME_SERVICE, &IID_IMFAudioStreamVolume, (void **)&as_volume)))
+    if (FAILED(MFGetService((IUnknown *)engine->session, &MR_POLICY_VOLUME_SERVICE, &IID_IMFSimpleAudioVolume, (void **)&sa_volume)))
         return;
 
-    if (FAILED(hr = IMFAudioStreamVolume_GetChannelCount(as_volume, &count)))
-        WARN("Failed to get channel count, hr %#lx.\n", hr);
-    for (i = 0; i < count; i++)
-    {
-        if (FAILED(hr = IMFAudioStreamVolume_SetChannelVolume(as_volume, i, engine->volume)))
-            WARN("Failed to set volume, hr %#lx.\n", hr);
-    }
-
-    IMFAudioStreamVolume_Release(as_volume);
+    if (FAILED(hr = IMFSimpleAudioVolume_SetMasterVolume(sa_volume, engine->volume)))
+        WARN("Failed to set master volume, hr %#lx.\n", hr);
+
+    IMFSimpleAudioVolume_Release(sa_volume);
 }
 
 static HRESULT WINAPI media_engine_callback_QueryInterface(IMFAsyncCallback *iface, REFIID riid, void **obj)
@@ -953,32 +936,19 @@
         case MESessionTopologyStatus:
         {
             UINT32 topo_status = 0;
-            IMFTopology *topology;
             PROPVARIANT value;
 
             IMFMediaEvent_GetUINT32(event, &MF_EVENT_TOPOLOGY_STATUS, &topo_status);
             if (topo_status != MF_TOPOSTATUS_READY)
                 break;
 
-            value.vt = VT_EMPTY;
-            if (FAILED(IMFMediaEvent_GetValue(event, &value)))
-                break;
-
-            if (value.vt != VT_UNKNOWN)
-            {
-                PropVariantClear(&value);
-                break;
-            }
-
-            topology = (IMFTopology *)value.punkVal;
-
             EnterCriticalSection(&engine->cs);
 
             media_engine_apply_volume(engine);
 
             engine->ready_state = MF_MEDIA_ENGINE_READY_HAVE_METADATA;
 
-            media_engine_get_frame_size(engine, topology);
+            media_engine_get_frame_size(engine);
 
             IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_DURATIONCHANGE, 0, 0);
             IMFMediaEngineNotify_EventNotify(engine->callback, MF_MEDIA_ENGINE_EVENT_LOADEDMETADATA, 0, 0);
@@ -1159,15 +1129,12 @@
     unsigned int category, role;
     IMFActivate *sar_activate;
     HRESULT hr;
-    GUID guid;
 
     *node = NULL;
 
     if (FAILED(hr = MFCreateAudioRendererActivate(&sar_activate)))
         return hr;
 
-    CoCreateGuid(&guid);
-    IMFActivate_SetGUID(sar_activate, &MF_AUDIO_RENDERER_ATTRIBUTE_SESSION_ID, &guid);
     /* Configuration attributes keys differ between Engine and SAR. */
     if (SUCCEEDED(IMFAttributes_GetUINT32(engine->attributes, &MF_MEDIA_ENGINE_AUDIO_CATEGORY, &category)))
         IMFActivate_SetUINT32(sar_activate, &MF_AUDIO_RENDERER_ATTRIBUTE_STREAM_CATEGORY, category);
@@ -1198,6 +1165,23 @@
     {
         WARN("Output format was not specified.\n");
         return E_FAIL;
+    }
+
+    switch (output_format)
+    {
+        case DXGI_FORMAT_R10G10B10A2_TYPELESS:
+        case DXGI_FORMAT_R10G10B10A2_UNORM:
+        case DXGI_FORMAT_R10G10B10A2_UINT:
+            /* IMFMediaSession doesn't support output to these formats unless the decoder supports
+             * MFVideoFormat_P010 output, which would allow inclusion of a suitable converter.
+             * The Windows H.264 decoder doesn't suppport MFVideoFormat_P010 output, and Media
+             * Engine apparently performs a format conversion.
+             * Create an 8-bit output and ensure the sampled texture is copied via a pixel shader. */
+            output_format = DXGI_FORMAT_B8G8R8A8_UNORM;
+            engine->video_frame.format_mismatch = TRUE;
+            break;
+        default:
+            break;
     }
 
     memcpy(&subtype, &MFVideoFormat_Base, sizeof(subtype));
@@ -1409,7 +1393,6 @@
         }
 
         IMFTopology_SetUINT32(topology, &MF_TOPOLOGY_ENUMERATE_SOURCE_TYPES, TRUE);
-        IMFTopology_SetUINT32(topology, &MF_TOPOLOGY_ENABLE_XVP_FOR_PLAYBACK, TRUE);
 
         if (SUCCEEDED(hr))
             hr = IMFMediaSession_SetTopology(engine->session, MFSESSION_SETTOPOLOGY_IMMEDIATE, topology);
@@ -1889,6 +1872,23 @@
         return S_OK;
     }
 
+    /* HACK: Don't seek if the time delta is too small. */
+    do
+    {
+        const char *game_id = getenv("SteamGameId");
+        MFTIME clocktime;
+
+        if (game_id && !strcmp(game_id, "3185890"))
+        {
+            if (IMFMediaEngineEx_IsPaused(&engine->IMFMediaEngineEx_iface)
+                    || FAILED(IMFPresentationClock_GetTime(engine->clock, &clocktime)))
+                break;
+
+            if (fabs(mftime_to_seconds(clocktime) - seektime) < 0.01)
+                return S_OK;
+        }
+    } while(0);
+
     engine->next_seek = NAN;
 
     position.vt = VT_I8;
@@ -2442,11 +2442,39 @@
     }
 }
 
+static HRESULT get_d3d11_resource_from_sample(IMFSample *sample, ID3D11Texture2D **resource, UINT *subresource)
+{
+    IMFDXGIBuffer *dxgi_buffer;
+    IMFMediaBuffer *buffer;
+    HRESULT hr;
+
+    *resource = NULL;
+    *subresource = 0;
+
+    if (FAILED(hr = IMFSample_GetBufferByIndex(sample, 0, &buffer)))
+        return hr;
+
+    if (SUCCEEDED(hr = IMFMediaBuffer_QueryInterface(buffer, &IID_IMFDXGIBuffer, (void **)&dxgi_buffer)))
+    {
+        IMFDXGIBuffer_GetSubresourceIndex(dxgi_buffer, subresource);
+        hr = IMFDXGIBuffer_GetResource(dxgi_buffer, &IID_ID3D11Texture2D, (void **)resource);
+        IMFDXGIBuffer_Release(dxgi_buffer);
+    }
+
+    IMFMediaBuffer_Release(buffer);
+    return hr;
+}
+
 static void media_engine_update_d3d11_frame_surface(ID3D11DeviceContext *context, struct media_engine *engine)
 {
     D3D11_TEXTURE2D_DESC surface_desc;
+    D3D11_TEXTURE2D_DESC src_desc;
     IMFMediaBuffer *media_buffer;
+    ID3D11Texture2D *src_texture;
+    ID3D11Device *device;
     IMFSample *sample;
+    UINT subresource;
+    HRESULT hr;
 
     if (!video_frame_sink_get_sample(engine->presentation.frame_sink, &sample))
         return;
@@ -2464,7 +2492,23 @@
         surface_desc.Width = 0;
     }
 
-    if (SUCCEEDED(IMFSample_ConvertToContiguousBuffer(sample, &media_buffer)))
+    if (SUCCEEDED(hr = get_d3d11_resource_from_sample(sample, &src_texture, &subresource)))
+    {
+
+        ID3D11Texture2D_GetDesc(src_texture, &src_desc);
+
+        if (SUCCEEDED(hr = media_engine_lock_d3d_device(engine, &device)))
+        {
+            ID3D11Device_GetImmediateContext(device, &context);
+            ID3D11DeviceContext_CopyResource(context, (ID3D11Resource *)engine->video_frame.d3d11.source, (ID3D11Resource *)src_texture);
+            ID3D11DeviceContext_Release(context);
+            media_engine_unlock_d3d_device(engine, device);
+        }
+
+        ID3D11Texture2D_Release(src_texture);
+    }
+
+    if (FAILED(hr) && SUCCEEDED(IMFSample_ConvertToContiguousBuffer(sample, &media_buffer)))
     {
         BYTE *buffer;
         DWORD buffer_size;
@@ -2482,29 +2526,6 @@
     }
 
     IMFSample_Release(sample);
-}
-
-static HRESULT get_d3d11_resource_from_sample(IMFSample *sample, ID3D11Texture2D **resource, UINT *subresource)
-{
-    IMFDXGIBuffer *dxgi_buffer;
-    IMFMediaBuffer *buffer;
-    HRESULT hr;
-
-    *resource = NULL;
-    *subresource = 0;
-
-    if (FAILED(hr = IMFSample_GetBufferByIndex(sample, 0, &buffer)))
-        return hr;
-
-    if (SUCCEEDED(hr = IMFMediaBuffer_QueryInterface(buffer, &IID_IMFDXGIBuffer, (void **)&dxgi_buffer)))
-    {
-        IMFDXGIBuffer_GetSubresourceIndex(dxgi_buffer, subresource);
-        hr = IMFDXGIBuffer_GetResource(dxgi_buffer, &IID_ID3D11Texture2D, (void **)resource);
-        IMFDXGIBuffer_Release(dxgi_buffer);
-    }
-
-    IMFMediaBuffer_Release(buffer);
-    return hr;
 }
 
 static HRESULT media_engine_transfer_d3d11(struct media_engine *engine, ID3D11Texture2D *dst_texture,
@@ -2849,7 +2870,9 @@
 
     if (SUCCEEDED(IUnknown_QueryInterface(surface, &IID_ID3D11Texture2D, (void **)&texture)))
     {
-        if (!engine->device_manager || FAILED(hr = media_engine_transfer_d3d11(engine, texture, src_rect, dst_rect, color)))
+        if (!engine->device_manager
+                || engine->video_frame.format_mismatch
+                || FAILED(hr = media_engine_transfer_d3d11(engine, texture, src_rect, dst_rect, color)))
             hr = media_engine_transfer_to_d3d11_texture(engine, texture, src_rect, dst_rect, color);
         ID3D11Texture2D_Release(texture);
     }
