diff --git a/dlls/wintypes/tests/wintypes.c b/dlls/wintypes/tests/wintypes.c
--- a/dlls/wintypes/tests/wintypes.c
+++ b/dlls/wintypes/tests/wintypes.c
@@ -28,9 +28,7 @@
 #include "rometadataresolution.h"
 
 #define WIDL_using_Windows_Foundation
-#define WIDL_using_Windows_Foundation_Collections
 #define WIDL_using_Windows_Foundation_Metadata
-#include "windows.foundation.h"
 #include "windows.foundation.metadata.h"
 #include "wintypes_test.h"
 
@@ -738,40 +736,15 @@
     IInspectable *inspectable = NULL, *tmp_inspectable = NULL;
     IPropertyValueStatics *statics = NULL;
     IActivationFactory *factory = NULL;
-    IReference_BYTE *iref_byte;
-    IReference_INT16 *iref_int16;
-    IReference_INT32 *iref_int32;
-    IReference_UINT32 *iref_uint32;
-    IReference_INT64 *iref_int64;
-    IReference_UINT64 *iref_uint64;
     IReference_boolean *iref_boolean;
     IReference_HSTRING *iref_hstring;
-    IReference_FLOAT *iref_float;
     IReference_DOUBLE *iref_double;
-    IReference_DateTime *iref_datetime;
-    IReference_TimeSpan *iref_timespan;
-    IReference_GUID *iref_guid;
-    IReference_Point *iref_point;
-    IReference_Size *iref_size;
-    IReference_Rect *iref_rect;
     IPropertyValue *value = NULL;
     enum PropertyType type;
     unsigned int i, count;
     BYTE byte, *ptr_byte;
     HSTRING str, ret_str;
-    INT16 ret_int16;
-    INT32 ret_int32;
-    UINT32 ret_uint32;
-    INT64 ret_int64;
-    UINT64 ret_uint64;
-    FLOAT ret_float;
     DOUBLE ret_double;
-    struct DateTime ret_datetime;
-    struct TimeSpan ret_timespan;
-    GUID ret_guid;
-    struct Point ret_point;
-    struct Size ret_size;
-    struct Rect ret_rect;
     boolean ret;
     HRESULT hr;
 
@@ -833,9 +806,6 @@
 
     hr = IPropertyValue_GetBoolean(value, &ret);
     ok(hr == TYPE_E_TYPEMISMATCH, "Got unexpected hr %#lx.\n", hr);
-
-    hr = IPropertyValueStatics_CreateString(statics, (HSTRING)-1, NULL);
-    ok(hr == E_POINTER, "Got unexpected hr %#lx.\n", hr);
 
     IPropertyValue_Release(value);
 
@@ -1019,29 +989,16 @@
                                                                                              \
         hr = IFACE_TYPE##_get_Value(RET_OBJ, &RET_VALUE);                                    \
         ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);                                     \
-        ok(!memcmp(&RET_VALUE, &VALUE, sizeof(VALUE)), "Got unexpected value.\n");           \
+        ok(RET_VALUE == VALUE, "Got unexpected value.\n");                                   \
                                                                                              \
         IFACE_TYPE##_Release(RET_OBJ);                                                       \
         IPropertyValue_Release(value);                                                       \
         IInspectable_Release(inspectable);                                                   \
     } while (0);
 
-    TEST_PROPERTY_VALUE_IREFERENCE(UInt8, IReference_BYTE, byte_value, iref_byte, byte)
-    TEST_PROPERTY_VALUE_IREFERENCE(Int16, IReference_INT16, int16_value, iref_int16, ret_int16)
-    TEST_PROPERTY_VALUE_IREFERENCE(Int32, IReference_INT32, int32_value, iref_int32, ret_int32)
-    TEST_PROPERTY_VALUE_IREFERENCE(UInt32, IReference_UINT32, uint32_value, iref_uint32, ret_uint32)
-    TEST_PROPERTY_VALUE_IREFERENCE(Int64, IReference_INT64, int64_value, iref_int64, ret_int64)
-    TEST_PROPERTY_VALUE_IREFERENCE(UInt64, IReference_UINT64, uint64_value, iref_uint64, ret_uint64)
-    TEST_PROPERTY_VALUE_IREFERENCE(Boolean, IReference_boolean, boolean_value, iref_boolean, ret)
+    TEST_PROPERTY_VALUE_IREFERENCE(Boolean, IReference_boolean, TRUE, iref_boolean, ret)
     TEST_PROPERTY_VALUE_IREFERENCE(String, IReference_HSTRING, str, iref_hstring, ret_str)
-    TEST_PROPERTY_VALUE_IREFERENCE(Single, IReference_FLOAT, float_value, iref_float, ret_float)
-    TEST_PROPERTY_VALUE_IREFERENCE(Double, IReference_DOUBLE, double_value, iref_double, ret_double)
-    TEST_PROPERTY_VALUE_IREFERENCE(DateTime, IReference_DateTime, datetime_value, iref_datetime, ret_datetime)
-    TEST_PROPERTY_VALUE_IREFERENCE(TimeSpan, IReference_TimeSpan, timespan_value, iref_timespan, ret_timespan)
-    TEST_PROPERTY_VALUE_IREFERENCE(Guid, IReference_GUID, IID_IPropertyValue, iref_guid, ret_guid)
-    TEST_PROPERTY_VALUE_IREFERENCE(Point, IReference_Point, point_value, iref_point, ret_point)
-    TEST_PROPERTY_VALUE_IREFERENCE(Size, IReference_Size, size_value, iref_size, ret_size)
-    TEST_PROPERTY_VALUE_IREFERENCE(Rect, IReference_Rect, rect_value, iref_rect, ret_rect)
+    TEST_PROPERTY_VALUE_IREFERENCE(Double, IReference_DOUBLE, 1.5, iref_double, ret_double)
 
 #undef TEST_PROPERTY_VALUE_IREFERENCE
 
@@ -1255,239 +1212,6 @@
     }
 }
 
-static void test_IPropertySet(void)
-{
-    static const WCHAR *class_name = RuntimeClass_Windows_Foundation_Collections_PropertySet;
-    IActivationFactory *propset_factory;
-    IPropertyValueStatics *propval_statics;
-    IInspectable *inspectable, *val, *val1, *val2, *val3;
-    IPropertySet *propset;
-    IMap_HSTRING_IInspectable *map;
-    IMapView_HSTRING_IInspectable *map_view;
-    IObservableMap_HSTRING_IInspectable *observable_map;
-    IIterable_IKeyValuePair_HSTRING_IInspectable *iterable;
-    IIterator_IKeyValuePair_HSTRING_IInspectable *iterator;
-    IKeyValuePair_HSTRING_IInspectable *pair;
-    BOOLEAN boolean;
-    HRESULT hr;
-    HSTRING name, key1, key2;
-    IPropertyValue *propval;
-    UINT32 uint32 = 0;
-
-    hr = RoInitialize( RO_INIT_MULTITHREADED );
-    ok( hr == S_OK, "got %#lx\n", hr );
-
-    hr = WindowsCreateString( class_name, wcslen( class_name ), &name );
-    ok( hr == S_OK, "got %#lx\n", hr );
-    hr = RoGetActivationFactory( name, &IID_IActivationFactory, (void **)&propset_factory );
-    WindowsDeleteString( name );
-    ok( hr == S_OK || broken( hr == REGDB_E_CLASSNOTREG ), "RoGetActivationFactory failed, hr %#lx.\n", hr );
-    if (hr != S_OK)
-    {
-        win_skip( "%s runtimeclass not registered, skipping tests.\n", wine_dbgstr_w( class_name ) );
-        goto done;
-    }
-
-    class_name = RuntimeClass_Windows_Foundation_PropertyValue;
-    hr = WindowsCreateString( class_name, wcslen( class_name ), &name );
-    ok( hr == S_OK, "got %#lx\n", hr );
-    hr = RoGetActivationFactory( name, &IID_IPropertyValueStatics, (void **)&propval_statics);
-    WindowsDeleteString( name );
-    ok( hr == S_OK || broken( hr == REGDB_E_CLASSNOTREG ), "RoGetActivationFactory failed, hr %#lx.\n", hr );
-    if (hr != S_OK)
-    {
-        win_skip( "%s runtimeclass not registered, skipping tests.\n", wine_dbgstr_w( class_name ) );
-        IActivationFactory_Release( propset_factory );
-        goto done;
-    }
-
-    hr = IActivationFactory_ActivateInstance( propset_factory, &inspectable );
-    IActivationFactory_Release( propset_factory );
-    ok( hr == S_OK, "got %#lx\n", hr );
-
-    hr = IInspectable_QueryInterface( inspectable, &IID_IPropertySet, (void **)&propset );
-    IInspectable_Release( inspectable );
-    ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-
-    hr = IPropertySet_QueryInterface( propset, &IID_IObservableMap_HSTRING_IInspectable, (void **)&observable_map );
-    ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-    IObservableMap_HSTRING_IInspectable_Release( observable_map );
-
-    hr = IPropertySet_QueryInterface( propset, &IID_IMap_HSTRING_IInspectable, (void **)&map );
-    IPropertySet_Release( propset );
-    ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-
-    boolean = TRUE;
-    hr = IMap_HSTRING_IInspectable_HasKey( map, NULL, &boolean );
-    ok( hr == S_OK, "HasKey failed, got %#lx\n", hr );
-    ok( !boolean, "Got boolean %d.\n", boolean );
-    hr = IMap_HSTRING_IInspectable_Lookup( map, NULL, &val );
-    ok( hr == E_BOUNDS, "Got hr %#lx\n", hr );
-
-    hr = IPropertyValueStatics_CreateUInt32( propval_statics, 0xdeadbeef, &val1 );
-    ok( hr == S_OK, "CreateUInt32 failed, got %#lx\n", hr );
-    boolean = TRUE;
-    hr = IMap_HSTRING_IInspectable_Insert( map, NULL, val1, &boolean );
-    ok( hr == S_OK, "Insert failed, got %#lx\n", hr );
-    ok( !boolean, "Got boolean %d.\n", boolean );
-
-    hr = IPropertyValueStatics_CreateUInt32( propval_statics, 0xc0decafe, &val2 );
-    ok( hr == S_OK, "CreateUInt32 failed, got %#lx\n", hr );
-    boolean = FALSE;
-    hr = IMap_HSTRING_IInspectable_Insert( map, NULL, val2, &boolean );
-    IInspectable_Release( val2 );
-    ok( boolean, "Got boolean %d.\n", boolean );
-    ok( hr == S_OK, "Insert failed, got %#lx\n", hr );
-    boolean = FALSE;
-    hr = IMap_HSTRING_IInspectable_HasKey( map, NULL, &boolean );
-    ok( hr == S_OK, "HasKey failed, got %#lx\n", hr );
-    ok( boolean, "Got boolean %d.\n", boolean );
-    hr = IMap_HSTRING_IInspectable_Lookup( map, NULL, &val );
-    ok( hr == S_OK, "Lookup failed, got %#lx\n", hr );
-    hr = IInspectable_QueryInterface( val, &IID_IPropertyValue, (void **)&propval );
-    IInspectable_Release( val );
-    ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-    hr = IPropertyValue_GetUInt32( propval, &uint32 );
-    IPropertyValue_Release( propval );
-    ok( hr == S_OK, "GetUInt32, got %#lx\n", hr );
-    ok( uint32 == 0xc0decafe, "Got uint32 %u\n", uint32 );
-
-    hr = WindowsCreateString( L"foo", 3, &key1 );
-    ok( hr == S_OK, "WindowsCreateString failed, got %#lx\n", hr );
-    boolean = TRUE;
-    hr = IMap_HSTRING_IInspectable_Lookup( map, key1, &val );
-    ok( hr == E_BOUNDS, "Got hr %#lx\n", hr );
-    hr = IMap_HSTRING_IInspectable_Insert( map, key1, val1, &boolean );
-    IInspectable_Release( val1 );
-    ok( hr == S_OK, "Insert failed, got %#lx\n", hr );
-    ok( !boolean, "Got boolean %d.\n", boolean );
-    boolean = FALSE;
-    hr = IMap_HSTRING_IInspectable_HasKey( map, key1, &boolean );
-    ok( hr == S_OK, "HasKey failed, got %#lx\n", hr );
-    ok( boolean, "Got boolean %d.\n", boolean );
-    hr = IMap_HSTRING_IInspectable_Lookup( map, key1, &val );
-    ok( hr == S_OK, "Lookup failed, got %#lx\n", hr );
-    hr = IInspectable_QueryInterface( val, &IID_IPropertyValue, (void **)&propval );
-    IInspectable_Release( val );
-    ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-    hr = IPropertyValue_GetUInt32( propval, &uint32 );
-    IPropertyValue_Release( propval );
-    ok( hr == S_OK, "GetUInt32, got %#lx\n", hr );
-    ok( uint32 == 0xdeadbeef, "Got uint32 %u\n", uint32 );
-    WindowsDeleteString( key1 );
-
-    hr = WindowsCreateString( L"bar", 3, &key2 );
-    ok( hr == S_OK, "WindowsCreateString failed, got %#lx\n", hr );
-    boolean = TRUE;
-    hr = IMap_HSTRING_IInspectable_HasKey( map, key2, &boolean );
-    ok( hr == S_OK, "HasKey failed, got %#lx\n", hr );
-    ok( !boolean, "Got boolean %d.\n", boolean );
-    hr = IPropertyValueStatics_CreateUInt64( propval_statics, 0xdeadbeefdeadbeef, &val3 );
-    ok( hr == S_OK, "CreateUInt32 failed, got %#lx\n", hr );
-    boolean = TRUE;
-    hr = IMap_HSTRING_IInspectable_Insert( map, key2, val3, &boolean );
-    IInspectable_Release( val3 );
-    ok( hr == S_OK, "Insert failed, got %#lx\n", hr );
-    ok( !boolean, "Got boolean %d.\n", boolean );
-    boolean = FALSE;
-    hr = IMap_HSTRING_IInspectable_HasKey( map, key2, &boolean );
-    ok( hr == S_OK, "HasKey failed, got %#lx\n", hr );
-    ok( boolean, "Got boolean %d.\n", boolean );
-    hr = IMap_HSTRING_IInspectable_Lookup( map, key2, &val );
-    ok( hr == S_OK, "Lookup failed, got %#lx\n", hr );
-    if (SUCCEEDED(hr))
-    {
-        IPropertyValue *propval;
-        UINT64 uint64 = 0;
-
-        hr = IInspectable_QueryInterface( val, &IID_IPropertyValue, (void **)&propval );
-        IInspectable_Release( val );
-        ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-        hr = IPropertyValue_GetUInt64( propval, &uint64 );
-        IPropertyValue_Release( propval );
-        ok( hr == S_OK, "GetUInt32, got %#lx\n", hr );
-        ok( uint64 == 0xdeadbeefdeadbeef, "Got uint64 %I64u\n", uint64 );
-    }
-
-    check_interface( map, &IID_IAgileObject, TRUE );
-    hr = IMap_HSTRING_IInspectable_QueryInterface( map, &IID_IIterable_IKeyValuePair_HSTRING_IInspectable,
-                                                   (void **)&iterable );
-    ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-    hr = IIterable_IKeyValuePair_HSTRING_IInspectable_First( iterable, &iterator );
-    ok( hr == S_OK, "got %#lx\n", hr );
-    IIterable_IKeyValuePair_HSTRING_IInspectable_Release( iterable );
-
-    check_interface( iterator, &IID_IAgileObject, TRUE );
-    hr = IIterator_IKeyValuePair_HSTRING_IInspectable_get_HasCurrent( iterator, &boolean );
-    ok( hr == S_OK, "Got hr %#lx\n", hr );
-    ok( boolean == TRUE, "Got %u\n", boolean );
-
-    hr = IIterator_IKeyValuePair_HSTRING_IInspectable_get_Current( iterator, &pair );
-    ok( hr == S_OK, "Got hr %#lx\n", hr );
-    check_interface( pair, &IID_IAgileObject, TRUE );
-    hr = IKeyValuePair_HSTRING_IInspectable_get_Key( pair, &key2 );
-    ok( hr == S_OK, "Got %#lx\n", hr );
-    WindowsDeleteString( key2 );
-    hr = IKeyValuePair_HSTRING_IInspectable_get_Value( pair, &val );
-    ok( hr == S_OK, "Got %#lx\n", hr );
-    IInspectable_Release( val );
-
-    hr = IMap_HSTRING_IInspectable_GetView( map, &map_view );
-    ok( hr == S_OK, "GetView failed, got %#lx\n", hr );
-
-    check_interface( map_view, &IID_IAgileObject, TRUE );
-    hr = IMapView_HSTRING_IInspectable_QueryInterface( map_view, &IID_IIterable_IKeyValuePair_HSTRING_IInspectable,
-                                                       (void **)&iterable );
-    ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-    hr = IMapView_HSTRING_IInspectable_Lookup( map_view, key2, &val );
-    ok( hr == S_OK, "Lookup failed, got %#lx\n", hr );
-    IInspectable_Release( val );
-
-
-    /* after map is modified, associated objects are invalidated */
-    hr = IMap_HSTRING_IInspectable_Remove( map, key2 );
-    ok( hr == S_OK, "Remove failed, got %#lx\n", hr );
-
-    hr = IKeyValuePair_HSTRING_IInspectable_get_Key( pair, &key2 );
-    ok( hr == S_OK, "Got %#lx\n", hr );
-    WindowsDeleteString( key2 );
-    hr = IKeyValuePair_HSTRING_IInspectable_get_Value( pair, &val );
-    ok( hr == S_OK, "Got %#lx\n", hr );
-    IInspectable_Release( val );
-    IKeyValuePair_HSTRING_IInspectable_Release( pair );
-
-    hr = IIterator_IKeyValuePair_HSTRING_IInspectable_get_HasCurrent( iterator, &boolean );
-    ok( hr == S_OK, "Got hr %#lx\n", hr );
-    ok( boolean == TRUE, "Got %u\n", boolean );
-    hr = IIterator_IKeyValuePair_HSTRING_IInspectable_get_Current( iterator, &pair );
-    ok( hr == E_CHANGED_STATE, "Got hr %#lx\n", hr );
-    IIterator_IKeyValuePair_HSTRING_IInspectable_Release( iterator );
-
-    hr = IMapView_HSTRING_IInspectable_Lookup( map_view, key2, &val );
-    ok( hr == E_CHANGED_STATE, "Got hr %#lx\n", hr );
-    WindowsDeleteString( key2 );
-    hr = IIterable_IKeyValuePair_HSTRING_IInspectable_First( iterable, &iterator );
-    ok( hr == E_CHANGED_STATE, "got %#lx\n", hr );
-    IIterable_IKeyValuePair_HSTRING_IInspectable_Release( iterable );
-    IMapView_HSTRING_IInspectable_Release( map_view );
-
-    hr = IMap_HSTRING_IInspectable_GetView( map, &map_view );
-    ok( hr == S_OK, "GetView failed, got %#lx\n", hr );
-    hr = IMapView_HSTRING_IInspectable_QueryInterface( map_view, &IID_IIterable_IKeyValuePair_HSTRING_IInspectable,
-                                                       (void **)&iterable );
-    ok( hr == S_OK, "QueryInterface failed, got %#lx\n", hr );
-    hr = IIterable_IKeyValuePair_HSTRING_IInspectable_First( iterable, &iterator );
-    ok( hr == S_OK, "got %#lx\n", hr );
-    IIterator_IKeyValuePair_HSTRING_IInspectable_Release( iterator );
-    IIterable_IKeyValuePair_HSTRING_IInspectable_Release( iterable );
-    IMapView_HSTRING_IInspectable_Release( map_view );
-
-    IMap_HSTRING_IInspectable_Release( map );
-    IPropertyValueStatics_Release( propval_statics );
-done:
-    RoUninitialize();
-}
-
 START_TEST(wintypes)
 {
     IsWow64Process(GetCurrentProcess(), &is_wow64);
@@ -1498,5 +1222,4 @@
     test_IPropertyValueStatics();
     test_RoParseTypeName();
     test_RoResolveNamespace();
-    test_IPropertySet();
 }
