diff --git a/dlls/dplayx/tests/dplayx.c b/dlls/dplayx/tests/dplayx.c
--- a/dlls/dplayx/tests/dplayx.c
+++ b/dlls/dplayx/tests/dplayx.c
@@ -838,7 +838,7 @@
     return hr;
 }
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 
 typedef struct
 {
@@ -1008,7 +1008,7 @@
     DWORD dataOffset;
 } GroupDataChanged;
 
-#pragma pack(pop)
+#include "poppack.h"
 
 #define bindUdp( port ) bindUdp_( __LINE__, port )
 static SOCKET bindUdp_( int line, unsigned short port )
@@ -1254,14 +1254,14 @@
 static unsigned short receiveEnumSessionsRequest_( int line, SOCKET sock, const GUID *expectedAppGuid,
                                                    const WCHAR *expectedPassword, DWORD expectedFlags )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
         EnumSessionsRequest request;
         WCHAR password[ 256 ];
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     DWORD expectedPasswordSize;
     unsigned short port;
     DWORD expectedSize;
@@ -1299,7 +1299,7 @@
 #define sendEnumSessionsReply( sock, port, dpsd ) sendEnumSessionsReply_( __LINE__, sock, port, dpsd )
 static void sendEnumSessionsReply_( int line, SOCKET sock, unsigned short port, const DPSESSIONDESC2 *dpsd )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -1328,7 +1328,7 @@
             .nameOffset = sizeof( reply.reply ),
         },
     };
-#pragma pack(pop)
+#include "poppack.h"
     DWORD passwordSize;
     int wsResult;
     DWORD size;
@@ -1348,13 +1348,13 @@
 #define receiveRequestPlayerId( sock, expectedFlags ) receiveRequestPlayerId_( __LINE__, sock, expectedFlags )
 static unsigned short receiveRequestPlayerId_( int line, SOCKET sock, DWORD expectedFlags )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
         RequestPlayerId request;
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     unsigned short port;
     int wsResult;
 
@@ -1371,7 +1371,7 @@
 #define sendRequestPlayerReply( sock, port, id, result ) sendRequestPlayerReply_( __LINE__, sock, port, id, result )
 static void sendRequestPlayerReply_( int line, SOCKET sock, unsigned short port, DPID id, HRESULT result )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -1399,7 +1399,7 @@
             .result = result,
         },
     };
-#pragma pack(pop)
+#include "poppack.h"
     int wsResult;
 
     wsResult = send( sock, (char *) &reply, sizeof( reply ), 0 );
@@ -1412,13 +1412,13 @@
                                                  const WCHAR *expectedPassword, DWORD expectedTickCount,
                                                  unsigned short *udpPort )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
         AddForwardRequest request;
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     DWORD expectedPasswordSize;
     WCHAR password[ 256 ];
     unsigned short port;
@@ -1467,7 +1467,7 @@
 {
 #define SHORT_NAME L"short name"
 #define LONG_NAME L"long name"
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -1678,7 +1678,7 @@
             .playerIds = { 0xd00de, },
         },
     };
-#pragma pack(pop)
+#include "poppack.h"
 #undef LONG_NAME
 #undef SHORT_NAME
     int wsResult;
@@ -1695,7 +1695,7 @@
 #define sendAddForwardReply( sock, port, result ) sendAddForwardReply_( __LINE__, sock, port, result )
 static void sendAddForwardReply_( int line, SOCKET sock, unsigned short port, HRESULT result )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -1722,7 +1722,7 @@
             .result = result,
         },
     };
-#pragma pack(pop)
+#include "poppack.h"
     int wsResult;
 
     wsResult = send( sock, (char *) &reply, sizeof( reply ), 0 );
@@ -1732,7 +1732,7 @@
 #define sendAddForward( sock, port, tcpPort, udpPort ) sendAddForward_( __LINE__, sock, port, tcpPort, udpPort )
 static void sendAddForward_( int line, SOCKET sock, unsigned short port, unsigned short tcpPort, unsigned short udpPort )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -1790,7 +1790,7 @@
             },
         },
     };
-#pragma pack(pop)
+#include "poppack.h"
 
     int wsResult;
 
@@ -1801,13 +1801,13 @@
 #define receiveAddForwardAck( sock, expectedPlayerId ) receiveAddForwardAck_( __LINE__, sock, expectedPlayerId )
 static unsigned short receiveAddForwardAck_( int line, SOCKET sock, DPID expectedPlayerId )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
         AddForwardAck request;
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     unsigned short port;
     int wsResult;
 
@@ -1828,13 +1828,13 @@
                                             const WCHAR *expectedShortName, const WCHAR *expectedLongName,
                                             void *expectedPlayerData, DWORD expectedPlayerDataSize )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
         CreatePlayer request;
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     DWORD expectedShortNameSize;
     DWORD expectedLongNameSize;
     WCHAR shortName[ 256 ];
@@ -1917,7 +1917,7 @@
 static void sendCreatePlayer_( int line, SOCKET sock, unsigned short tcpPort, unsigned short udpPort,
                                const WCHAR *shortName, const WCHAR *longName, void *playerData, DWORD playerDataSize )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -1963,7 +1963,7 @@
             },
         },
     };
-#pragma pack(pop)
+#include "poppack.h"
     SpData spData = {
         .tcpAddr.sin_family = AF_INET,
         .tcpAddr.sin_port = htons( tcpPort ),
@@ -2021,14 +2021,14 @@
 static unsigned short receiveGuaranteedGameMessage_( int line, SOCKET sock, DPID expectedFromId, DPID expectedToId,
                                                      void *expectedData, DWORD expectedDataSize )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
         GameMessage request;
         BYTE data[ 256 ];
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     unsigned short port;
     int wsResult;
 
@@ -2049,13 +2049,13 @@
 static void receiveGameMessage_( int line, SOCKET sock, DPID expectedFromId, DPID expectedToId, void *expectedData,
                                  DWORD expectedDataSize )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         GameMessage request;
         BYTE data[ 256 ];
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     int wsResult;
 
     DWORD expectedSize = sizeof( request.request ) + expectedDataSize;
@@ -2072,7 +2072,7 @@
 static void sendGuaranteedGameMessage_( int line, SOCKET sock, unsigned short port, DPID fromId, DPID toId, void *data,
                                         DWORD dataSize )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -2094,7 +2094,7 @@
             .toId = toId,
         }
     };
-#pragma pack(pop)
+#include "poppack.h"
     int wsResult;
 
     wsResult = send( sock, (char *) &request, sizeof( request ), 0 );
@@ -2108,7 +2108,7 @@
         sendGameMessage_( __LINE__, sock, fromId, toId, data, dataSize )
 static void sendGameMessage_( int line, SOCKET sock, DPID fromId, DPID toId, void *data, DWORD dataSize )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         GameMessage request;
@@ -2121,7 +2121,7 @@
             .toId = toId,
         }
     };
-#pragma pack(pop)
+#include "poppack.h"
     int wsResult;
     DWORD size;
 
@@ -2137,7 +2137,7 @@
         sendPing_( __LINE__, sock, port, fromId, tickCount )
 static void sendPing_( int line, SOCKET sock, unsigned short port, DPID fromId, DWORD tickCount )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -2165,7 +2165,7 @@
             .tickCount = tickCount,
         }
     };
-#pragma pack(pop)
+#include "poppack.h"
     int wsResult;
 
     wsResult = send( sock, (char *) &request, sizeof( request ), 0 );
@@ -2176,13 +2176,13 @@
         receivePingReply_( __LINE__, sock, expectedFromId, expectedTickCount )
 static unsigned short receivePingReply_( int line, SOCKET sock, DPID expectedFromId, DWORD expectedTickCount )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
         Ping request;
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     unsigned short port;
     int wsResult;
 
@@ -2202,13 +2202,13 @@
         receiveAddPlayerToGroup_( __LINE__, sock, expectedPlayerId, expectedGroupId )
 static unsigned short receiveAddPlayerToGroup_( int line, SOCKET sock, DPID expectedPlayerId, DPID expectedGroupId )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
         AddPlayerToGroup request;
     } request;
-#pragma pack(pop)
+#include "poppack.h"
     unsigned short port;
     int wsResult;
 
@@ -2233,7 +2233,7 @@
         sendGroupDataChanged_( __LINE__, sock, port, groupId, data, dataSize )
 static void sendGroupDataChanged_( int line, SOCKET sock, unsigned short port, DPID groupId, void *data, DWORD dataSize )
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct
     {
         SpHeader spHeader;
@@ -2264,7 +2264,7 @@
             .dataOffset = sizeof( request.request ),
         },
     };
-#pragma pack(pop)
+#include "poppack.h"
     int wsResult;
     int size;
 
