diff --git a/dlls/d3dx9_36/effect.c b/dlls/d3dx9_36/effect.c
--- a/dlls/d3dx9_36/effect.c
+++ b/dlls/d3dx9_36/effect.c
@@ -224,7 +224,7 @@
         struct IDirect3DDevice9 *device, const char *data, SIZE_T data_size,
         unsigned int flags, struct ID3DXEffectPool *pool, const char *skip_constants_string);
 static HRESULT d3dx_parse_state(struct d3dx_effect *effect, struct d3dx_state *state,
-        const char *data, const char **ptr);
+        const char *data, const char **ptr, struct d3dx_object *objects);
 static void free_parameter(struct d3dx_parameter *param, BOOL element, BOOL child);
 
 typedef BOOL (*walk_parameter_dep_func)(void *data, struct d3dx_parameter *param);
@@ -2454,7 +2454,6 @@
 {
     struct d3dx_effect *effect = impl_from_ID3DXEffect(iface);
     struct d3dx_parameter *param = get_valid_parameter(effect, parameter);
-    unsigned int i;
 
     TRACE("iface %p, parameter %p, data %p, bytes %u.\n", iface, parameter, data, bytes);
 
@@ -2476,21 +2475,11 @@
         switch (param->type)
         {
             case D3DXPT_VOID:
+            case D3DXPT_BOOL:
             case D3DXPT_INT:
             case D3DXPT_FLOAT:
             case D3DXPT_STRING:
                 break;
-
-            case D3DXPT_BOOL:
-            {
-                BOOL *src = (BOOL *)param->data;
-                BOOL *dst = (BOOL *)data;
-
-                for (i = 0; i < (param->bytes / sizeof(*src)); ++i)
-                    dst[i] = !!src[i];
-
-                return D3D_OK;
-            }
 
             case D3DXPT_VERTEXSHADER:
             case D3DXPT_PIXELSHADER:
@@ -2499,6 +2488,9 @@
             case D3DXPT_TEXTURE2D:
             case D3DXPT_TEXTURE3D:
             case D3DXPT_TEXTURECUBE:
+            {
+                unsigned int i;
+
                 for (i = 0; i < (param->element_count ? param->element_count : 1); ++i)
                 {
                     IUnknown *unk = ((IUnknown **)param->data)[i];
@@ -2506,6 +2498,7 @@
                         IUnknown_AddRef(unk);
                 }
                 break;
+            }
 
             default:
                 FIXME("Unhandled type %s.\n", debug_d3dxparameter_type(param->type));
@@ -4439,77 +4432,10 @@
 static HRESULT WINAPI d3dx_effect_SetRawValue(ID3DXEffect *iface, D3DXHANDLE parameter, const void *data,
         UINT byte_offset, UINT bytes)
 {
-    struct d3dx_effect *effect = impl_from_ID3DXEffect(iface);
-    struct d3dx_parameter *param = get_valid_parameter(effect, parameter);
-
-    TRACE("iface %p, parameter %p, data %p, byte_offset %u, bytes %u.\n", iface, parameter, data, byte_offset, bytes);
-
-    if (!param)
-    {
-        WARN("Invalid parameter %p specified.\n", parameter);
-        return D3DERR_INVALIDCALL;
-    }
-
-    switch (param->class)
-    {
-        case D3DXPC_VECTOR:
-        {
-            uint8_t *dst_data;
-
-            if (param->columns != 4)
-            {
-                FIXME("Vec%u parameters are currently unsupported.\n", param->columns);
-                return E_NOTIMPL;
-            }
-
-            if ((byte_offset + bytes) > param->bytes)
-            {
-                FIXME("Writing adjacent parameters is currently unsupported.\n");
-                return E_NOTIMPL;
-            }
-
-            dst_data = param_get_data_and_dirtify(effect, param, !byte_offset ? bytes : param->bytes, TRUE);
-            memcpy(dst_data + byte_offset, data, bytes);
-            break;
-        }
-
-        case D3DXPC_MATRIX_ROWS:
-        {
-            D3DXMATRIX *dst_elem;
-            uint8_t *dst_data;
-            unsigned int i;
-
-            if (param->columns != 4 || param->rows != 4)
-            {
-                FIXME("%ux%u matrix parameters are currently unsupported.\n", param->rows, param->columns);
-                return E_NOTIMPL;
-            }
-
-            if (byte_offset & 0x3f || bytes & 0x3f)
-            {
-                FIXME("Partial matrix updates are currently unsupported.\n");
-                return E_NOTIMPL;
-            }
-
-            if ((byte_offset + bytes) > param->bytes)
-            {
-                FIXME("Writing adjacent parameters is currently unsupported.\n");
-                return E_NOTIMPL;
-            }
-
-            dst_data = param_get_data_and_dirtify(effect, param, !byte_offset ? bytes : param->bytes, TRUE);
-            dst_elem = (D3DXMATRIX *)(dst_data + byte_offset);
-            for (i = 0; i < (bytes / sizeof(D3DXMATRIX)); ++i)
-                D3DXMatrixTranspose(&dst_elem[i], &((const D3DXMATRIX *)data)[i]);
-            break;
-        }
-
-        default:
-            FIXME("Unhandled parameter class %s.\n", debug_d3dxparameter_class(param->class));
-            return E_NOTIMPL;
-    }
-
-    return D3D_OK;
+    FIXME("iface %p, parameter %p, data %p, byte_offset %u, bytes %u stub!\n",
+            iface, parameter, data, byte_offset, bytes);
+
+    return E_NOTIMPL;
 }
 #endif
 
@@ -5175,7 +5101,7 @@
 };
 
 static HRESULT d3dx_parse_sampler(struct d3dx_effect *effect, struct d3dx_sampler *sampler,
-        const char *data, const char **ptr)
+        const char *data, const char **ptr, struct d3dx_object *objects)
 {
     HRESULT hr;
     UINT i;
@@ -5192,7 +5118,7 @@
 
     for (i = 0; i < sampler->state_count; ++i)
     {
-        hr = d3dx_parse_state(effect, &sampler->states[i], data, ptr);
+        hr = d3dx_parse_state(effect, &sampler->states[i], data, ptr, objects);
         if (hr != D3D_OK)
         {
             WARN("Failed to parse state %u\n", i);
@@ -5215,7 +5141,7 @@
 }
 
 static HRESULT d3dx_parse_value(struct d3dx_effect *effect, struct d3dx_parameter *param,
-        void *value, const char *data, const char **ptr)
+        void *value, const char *data, const char **ptr, struct d3dx_object *objects)
 {
     unsigned int i;
     HRESULT hr;
@@ -5229,7 +5155,7 @@
         {
             struct d3dx_parameter *member = &param->members[i];
 
-            hr = d3dx_parse_value(effect, member, value ? (char *)value + old_size : NULL, data, ptr);
+            hr = d3dx_parse_value(effect, member, value ? (char *)value + old_size : NULL, data, ptr, objects);
             if (hr != D3D_OK)
             {
                 WARN("Failed to parse value %u\n", i);
@@ -5258,7 +5184,7 @@
             {
                 struct d3dx_parameter *member = &param->members[i];
 
-                hr = d3dx_parse_value(effect, member, (char *)value + old_size, data, ptr);
+                hr = d3dx_parse_value(effect, member, (char *)value + old_size, data, ptr, objects);
                 if (hr != D3D_OK)
                 {
                     WARN("Failed to parse value %u\n", i);
@@ -5282,7 +5208,7 @@
                 case D3DXPT_VERTEXSHADER:
                     param->object_id = read_u32(ptr);
                     TRACE("Id: %u\n", param->object_id);
-                    effect->objects[param->object_id].param = param;
+                    objects[param->object_id].param = param;
                     param->data = value;
                     break;
 
@@ -5298,7 +5224,7 @@
                     if (!sampler)
                         return E_OUTOFMEMORY;
 
-                    hr = d3dx_parse_sampler(effect, sampler, data, ptr);
+                    hr = d3dx_parse_sampler(effect, sampler, data, ptr, objects);
                     if (hr != D3D_OK)
                     {
                         free(sampler);
@@ -5325,7 +5251,7 @@
 }
 
 static HRESULT d3dx_parse_init_value(struct d3dx_effect *effect, struct d3dx_parameter *param,
-        const char *data, const char *ptr)
+        const char *data, const char *ptr, struct d3dx_object *objects)
 {
     UINT size = param->bytes;
     HRESULT hr;
@@ -5362,7 +5288,7 @@
         }
     }
 
-    hr = d3dx_parse_value(effect, param, value, data, &ptr);
+    hr = d3dx_parse_value(effect, param, value, data, &ptr, objects);
     if (hr != D3D_OK)
     {
         WARN("Failed to parse value\n");
@@ -5699,7 +5625,7 @@
 }
 
 static HRESULT d3dx_parse_effect_annotation(struct d3dx_effect *effect, struct d3dx_parameter *anno,
-        const char *data, const char **ptr)
+        const char *data, const char **ptr, struct d3dx_object *objects)
 {
     const char *ptr2;
     uint32_t offset;
@@ -5719,7 +5645,7 @@
 
     offset = read_u32(ptr);
     TRACE("Value offset: %#x.\n", offset);
-    hr = d3dx_parse_init_value(effect, anno, data, data + offset);
+    hr = d3dx_parse_init_value(effect, anno, data, data + offset, objects);
     if (hr != D3D_OK)
     {
         WARN("Failed to parse value.\n");
@@ -5730,7 +5656,7 @@
 }
 
 static HRESULT d3dx_parse_state(struct d3dx_effect *effect, struct d3dx_state *state,
-        const char *data, const char **ptr)
+        const char *data, const char **ptr, struct d3dx_object *objects)
 {
     struct d3dx_parameter *param = &state->parameter;
     enum STATE_CLASS state_class;
@@ -5764,7 +5690,7 @@
 
     offset = read_u32(ptr);
     TRACE("Value offset: %#x.\n", offset);
-    hr = d3dx_parse_init_value(effect, param, data, data + offset);
+    hr = d3dx_parse_init_value(effect, param, data, data + offset, objects);
     if (hr != D3D_OK)
     {
         WARN("Failed to parse value\n");
@@ -5805,7 +5731,7 @@
 }
 
 static HRESULT d3dx_parse_effect_parameter(struct d3dx_effect *effect, struct d3dx_top_level_parameter *param,
-        const char *data, const char **ptr)
+        const char *data, const char **ptr, struct d3dx_object *objects)
 {
     const char *ptr2;
     uint32_t offset;
@@ -5832,7 +5758,7 @@
         return hr;
     }
 
-    hr = d3dx_parse_init_value(effect, &param->param, data, data + offset);
+    hr = d3dx_parse_init_value(effect, &param->param, data, data + offset, objects);
     if (hr != D3D_OK)
     {
         WARN("Failed to parse value.\n");
@@ -5852,7 +5778,7 @@
         for (i = 0; i < param->annotation_count; ++i)
         {
             param_set_magic_number(&param->annotations[i]);
-            hr = d3dx_parse_effect_annotation(effect, &param->annotations[i], data, ptr);
+            hr = d3dx_parse_effect_annotation(effect, &param->annotations[i], data, ptr, objects);
             if (hr != D3D_OK)
             {
                 WARN("Failed to parse annotation.\n");
@@ -5877,7 +5803,7 @@
 }
 
 static HRESULT d3dx_parse_effect_pass(struct d3dx_effect *effect, struct d3dx_pass *pass,
-        const char *data, const char **ptr)
+        const char *data, const char **ptr, struct d3dx_object *objects)
 {
     struct d3dx_state *states = NULL;
     char *name = NULL;
@@ -5913,7 +5839,7 @@
         for (i = 0; i < pass->annotation_count; ++i)
         {
             param_set_magic_number(&pass->annotations[i]);
-            hr = d3dx_parse_effect_annotation(effect, &pass->annotations[i], data, ptr);
+            hr = d3dx_parse_effect_annotation(effect, &pass->annotations[i], data, ptr, objects);
             if (hr != D3D_OK)
             {
                 WARN("Failed to parse annotation %u\n", i);
@@ -5934,7 +5860,7 @@
 
         for (i = 0; i < pass->state_count; ++i)
         {
-            hr = d3dx_parse_state(effect, &states[i], data, ptr);
+            hr = d3dx_parse_state(effect, &states[i], data, ptr, objects);
             if (hr != D3D_OK)
             {
                 WARN("Failed to parse annotation %u\n", i);
@@ -5973,7 +5899,7 @@
 }
 
 static HRESULT d3dx_parse_effect_technique(struct d3dx_effect *effect, struct d3dx_technique *technique,
-        const char *data, const char **ptr)
+        const char *data, const char **ptr, struct d3dx_object *objects)
 {
     char *name = NULL;
     uint32_t offset;
@@ -6008,7 +5934,7 @@
         for (i = 0; i < technique->annotation_count; ++i)
         {
             param_set_magic_number(&technique->annotations[i]);
-            hr = d3dx_parse_effect_annotation(effect, &technique->annotations[i], data, ptr);
+            hr = d3dx_parse_effect_annotation(effect, &technique->annotations[i], data, ptr, objects);
             if (hr != D3D_OK)
             {
                 WARN("Failed to parse annotation %u\n", i);
@@ -6029,7 +5955,7 @@
 
         for (i = 0; i < technique->pass_count; ++i)
         {
-            hr = d3dx_parse_effect_pass(effect, &technique->passes[i], data, ptr);
+            hr = d3dx_parse_effect_pass(effect, &technique->passes[i], data, ptr, objects);
             if (hr != D3D_OK)
             {
                 WARN("Failed to parse pass %u\n", i);
@@ -6129,6 +6055,7 @@
         FIXME("Referenced parameter %s not found.\n", ptr + 4);
         return D3DXERR_INVALIDDATA;
     }
+    TRACE("Unknown u32: 0x%.8x.\n", *(uint32_t *)(ptr + string_size));
 
     if (string_size % sizeof(uint32_t))
         FIXME("Unaligned string_size %u.\n", string_size);
@@ -6388,7 +6315,7 @@
     for (i = 0; i < effect->params.count; ++i)
     {
         param_set_magic_number(&effect->params.parameters[i].param);
-        hr = d3dx_parse_effect_parameter(effect, &effect->params.parameters[i], data, &ptr);
+        hr = d3dx_parse_effect_parameter(effect, &effect->params.parameters[i], data, &ptr, effect->objects);
         if (hr != D3D_OK)
         {
             WARN("Failed to parse parameter %u.\n", i);
@@ -6411,7 +6338,7 @@
         for (i = 0; i < effect->technique_count; ++i)
         {
             TRACE("Parsing technique %u.\n", i);
-            hr = d3dx_parse_effect_technique(effect, &effect->techniques[i], data, &ptr);
+            hr = d3dx_parse_effect_technique(effect, &effect->techniques[i], data, &ptr, effect->objects);
             if (hr != D3D_OK)
             {
                 WARN("Failed to parse technique %u.\n", i);
