diff --git a/dlls/setupapi/devinst.c b/dlls/setupapi/devinst.c
--- a/dlls/setupapi/devinst.c
+++ b/dlls/setupapi/devinst.c
@@ -38,6 +38,7 @@
 #include "shlwapi.h"
 #include "wine/debug.h"
 #include "wine/list.h"
+#include "objbase.h"
 #include "cfgmgr32.h"
 #include "winioctl.h"
 #include "rpc.h"
@@ -2368,6 +2369,13 @@
                                         &deviceClass);
                                 if ((device = create_device(set, &deviceClass, deviceInst, FALSE)))
                                     SETUPDI_AddDeviceInterfaces(device, subKey, guid, flags);
+                                /*
+                                 * If there weren't any interfaces present for
+                                 * this device, remove it from the device info
+                                 * list.
+                                 */
+                                if (device && ((flags & DIGCF_PRESENT) && !list_count(&device->interfaces)))
+                                    remove_device(device);
                             }
                             RegCloseKey(deviceKey);
                         }
@@ -2441,6 +2449,94 @@
     }
 }
 
+
+/* iterate over all interfaces supported by this device instance. if any of
+ * them are "linked", return TRUE */
+static BOOL is_device_instance_linked(HKEY subKey, HKEY interfacesKey, const WCHAR *deviceInstance)
+{
+    LONG l;
+    DWORD class_idx = 0, device_idx, len, type;
+    HKEY class_key, device_key, link_key;
+    WCHAR class_keyname[40], device_keyname[MAX_DEVICE_ID_LEN];
+    WCHAR interface_devinstance[MAX_DEVICE_ID_LEN];
+    WCHAR alt_devinstance[MAX_DEVICE_ID_LEN];
+    WCHAR service[40];
+
+    l = RegQueryValueExW(subKey, L"Service", NULL, &type, (BYTE *)service, &len);
+    if (!l && type == REG_SZ && !wcsicmp(service, L"winehid") && !wcsstr(deviceInstance, L"HID\\"))
+    {
+        const WCHAR *tmp = wcschr(deviceInstance, L'\\');
+
+        if (tmp && (swprintf(alt_devinstance, ARRAY_SIZE(alt_devinstance), L"HID%s", tmp) != -1))
+        {
+            TRACE("Found winehid device instance, checking for link with device instance %s.\n", debugstr_w(alt_devinstance));
+            deviceInstance = alt_devinstance;
+        }
+    }
+
+    while (1)
+    {
+        len = ARRAY_SIZE(class_keyname);
+        l = RegEnumKeyExW(interfacesKey, class_idx++, class_keyname, &len, NULL, NULL, NULL, NULL);
+        if (l)
+            break;
+
+        l = RegOpenKeyExW(interfacesKey, class_keyname, 0, KEY_READ, &class_key);
+        if (l)
+            continue;
+
+        device_idx = 0;
+        while (1)
+        {
+            len = ARRAY_SIZE(device_keyname);
+            l = RegEnumKeyExW(class_key, device_idx++, device_keyname, &len, NULL, NULL, NULL, NULL);
+            if (l)
+                break;
+
+            l = RegOpenKeyExW(class_key, device_keyname, 0, KEY_READ, &device_key);
+            if (l)
+                continue;
+
+            len = ARRAY_SIZE(interface_devinstance);
+            l = RegQueryValueExW(device_key, L"DeviceInstance", NULL, &type, (BYTE *)interface_devinstance, &len);
+            if (l || type != REG_SZ)
+            {
+                RegCloseKey(device_key);
+                continue;
+            }
+
+            if (lstrcmpiW(interface_devinstance, deviceInstance))
+            {
+                /* not our device instance */
+                RegCloseKey(device_key);
+                continue;
+            }
+
+            l = RegOpenKeyExW(device_key, L"#", 0, KEY_READ, &link_key);
+            if (l)
+            {
+                RegCloseKey(device_key);
+                continue;
+            }
+
+            if (is_linked(link_key))
+            {
+                RegCloseKey(link_key);
+                RegCloseKey(device_key);
+                RegCloseKey(class_key);
+                return TRUE;
+            }
+
+            RegCloseKey(link_key);
+            RegCloseKey(device_key);
+        }
+
+        RegCloseKey(class_key);
+    }
+
+    return FALSE;
+}
+
 static void SETUPDI_EnumerateMatchingDeviceInstances(struct DeviceInfoSet *set,
         LPCWSTR enumerator, LPCWSTR deviceName, HKEY deviceKey,
         const GUID *class, DWORD flags)
@@ -2449,6 +2545,7 @@
     DWORD i, len;
     WCHAR deviceInstance[MAX_PATH];
     LONG l = ERROR_SUCCESS;
+    HKEY interfacesKey = SetupDiOpenClassRegKeyExW(NULL, KEY_READ, DIOCR_INTERFACE, NULL, NULL);
 
     TRACE("%s %s\n", debugstr_w(enumerator), debugstr_w(deviceName));
 
@@ -2481,14 +2578,16 @@
                         if ((flags & DIGCF_ALLCLASSES) ||
                                 IsEqualGUID(class, &deviceClass))
                         {
-                            static const WCHAR fmt[] =
-                             {'%','s','\\','%','s','\\','%','s',0};
+                            static const WCHAR fmt[] = L"%s\\%s\\%s";
 
                             if (swprintf(id, ARRAY_SIZE(id), fmt, enumerator,
-                                    deviceName, deviceInstance) != -1)
+                                        deviceName, deviceInstance) != -1 &&
+                                    (!(flags & DIGCF_PRESENT) ||
+                                     is_device_instance_linked(subKey, interfacesKey, id)))
                             {
                                 create_device(set, &deviceClass, id, FALSE);
                             }
+
                         }
                     }
                 }
@@ -2498,6 +2597,8 @@
             l = ERROR_SUCCESS;
         }
     }
+
+    RegCloseKey(interfacesKey);
 }
 
 static void SETUPDI_EnumerateMatchingDevices(HDEVINFO DeviceInfoSet,
@@ -4612,7 +4713,7 @@
 {
     CONFIGRET ret;
     DEVINSTID_W device_idw;
-    unsigned int slen;
+    WCHAR *p;
 
     TRACE("%p %s %#lx %p.\n", devinst, debugstr_a(device_id), flags, machine);
 
@@ -4622,11 +4723,13 @@
         return CR_CALL_NOT_IMPLEMENTED;
     }
 
-    slen = strlen(device_id) + 1;
-    if (!(device_idw = malloc(slen * sizeof(*device_idw))))
+    if (!(device_idw = malloc((strlen(device_id) + 1) * sizeof(*device_idw))))
         return CR_OUT_OF_MEMORY;
 
-    MultiByteToWideChar(CP_ACP, 0, device_id, slen, device_idw, slen);
+    p = device_idw;
+    while ((*p++ = *device_id++))
+        ;
+
     ret = CM_Locate_DevNode_ExW(devinst, device_idw, flags, NULL);
     free(device_idw);
     return ret;
@@ -4674,8 +4777,8 @@
     GUID guid, *pguid = NULL;
     unsigned int i, id_len;
     ULONG query_flags = 0;
+    WCHAR id[256];
     HDEVINFO set;
-    WCHAR id[256];
     ULONG needed;
     WCHAR *p;
 
@@ -4720,7 +4823,7 @@
     p = buffer;
     for (i = 0; SetupDiEnumDeviceInfo(set, i, &device); ++i)
     {
-        ret = SetupDiGetDeviceInstanceIdW(set, &device, id, ARRAY_SIZE(id), NULL);
+        ret = SetupDiGetDeviceInstanceIdW(set, &device, id, sizeof(id) / sizeof(*id), NULL);
         if (!ret) continue;
         id_len = wcslen(id) + 1;
         needed += id_len;
@@ -4744,14 +4847,11 @@
 }
 
 /***********************************************************************
- *             CM_Get_Device_ID_List_ExW  (SETUPAPI.@)
- */
-CONFIGRET WINAPI CM_Get_Device_ID_List_ExW(const WCHAR *filter, WCHAR *buffer, ULONG len, ULONG flags, HMACHINE machine)
-{
-    TRACE("%s %p %ld %#lx %p.\n", debugstr_w(filter), buffer, len, flags, machine);
-
-    if (machine)
-        FIXME("machine %p.\n", machine);
+ *             CM_Get_Device_ID_ListW  (SETUPAPI.@)
+ */
+CONFIGRET WINAPI CM_Get_Device_ID_ListW(const WCHAR *filter, WCHAR *buffer, ULONG len, ULONG flags)
+{
+    TRACE("%s %p %ld %#lx.\n", debugstr_w(filter), buffer, len, flags);
 
     if (!buffer)
         return CR_INVALID_POINTER;
@@ -4760,49 +4860,24 @@
 }
 
 /***********************************************************************
- *             CM_Get_Device_ID_ListW  (SETUPAPI.@)
- */
-CONFIGRET WINAPI CM_Get_Device_ID_ListW(const WCHAR *filter, WCHAR *buffer, ULONG len, ULONG flags)
-{
-    return CM_Get_Device_ID_List_ExW(filter, buffer, len, flags, NULL);
-}
-
-/***********************************************************************
- *             CM_Get_Device_ID_List_Size_ExW  (SETUPAPI.@)
- */
-CONFIGRET WINAPI CM_Get_Device_ID_List_Size_ExW(ULONG *len, const WCHAR *filter, ULONG flags, HMACHINE machine)
-{
-    TRACE("%p %s %#lx, machine %p.\n", len, debugstr_w(filter), flags, machine);
-
-    if (machine)
-        FIXME("machine %p.\n", machine);
+ *             CM_Get_Device_ID_List_SizeW  (SETUPAPI.@)
+ */
+CONFIGRET WINAPI CM_Get_Device_ID_List_SizeW(ULONG *len, const WCHAR *filter, ULONG flags)
+{
+    TRACE("%p %s %#lx.\n", len, debugstr_w(filter), flags);
 
     return get_device_id_list(filter, NULL, len, flags);
 }
 
 /***********************************************************************
- *             CM_Get_Device_ID_List_SizeW  (SETUPAPI.@)
- */
-CONFIGRET WINAPI CM_Get_Device_ID_List_SizeW(ULONG *len, const WCHAR *filter, ULONG flags)
-{
-    TRACE("%p %s %#lx.\n", len, debugstr_w(filter), flags);
-
-    return get_device_id_list(filter, NULL, len, flags);
-}
-
-/***********************************************************************
- *             CM_Get_Device_ID_List_ExA  (SETUPAPI.@)
- */
-CONFIGRET WINAPI CM_Get_Device_ID_List_ExA(const char *filter, char *buffer, ULONG len, ULONG flags, HMACHINE machine)
+ *             CM_Get_Device_ID_ListA  (SETUPAPI.@)
+ */
+CONFIGRET WINAPI CM_Get_Device_ID_ListA(const char *filter, char *buffer, ULONG len, ULONG flags )
 {
     WCHAR *wbuffer, *wfilter = NULL, *p;
-    unsigned int slen;
     CONFIGRET ret;
 
     TRACE("%s %p %ld %#lx.\n", debugstr_a(filter), buffer, len, flags);
-
-    if (machine)
-        FIXME("machine %p.\n", machine);
 
     if (!buffer || !len)
         return CR_INVALID_POINTER;
@@ -4812,13 +4887,15 @@
 
     if (filter)
     {
-        slen = strlen(filter) + 1;
-        if (!(wfilter = malloc(slen * sizeof(*wfilter))))
+        if (!(wfilter = malloc((strlen(filter) + 1) * sizeof(*wfilter))))
         {
             free(wbuffer);
             return CR_OUT_OF_MEMORY;
         }
-        MultiByteToWideChar(CP_ACP, 0, filter, slen, wfilter, slen);
+
+        p = wfilter;
+        while ((*p++ = *filter++))
+            ;
     }
 
     if (!(ret = CM_Get_Device_ID_ListW(wfilter, wbuffer, len, flags)))
@@ -4826,10 +4903,8 @@
         p = wbuffer;
         while (*p)
         {
-            slen = wcslen(p) + 1;
-            WideCharToMultiByte(CP_ACP, 0, p, slen, buffer, slen, NULL, NULL);
-            p += slen;
-            buffer += slen;
+            while ((*buffer++ = *p++))
+                ;
         }
         *buffer = 0;
     }
@@ -4839,45 +4914,27 @@
 }
 
 /***********************************************************************
- *             CM_Get_Device_ID_ListA  (SETUPAPI.@)
- */
-CONFIGRET WINAPI CM_Get_Device_ID_ListA(const char *filter, char *buffer, ULONG len, ULONG flags)
-{
-    return CM_Get_Device_ID_List_ExA(filter, buffer, len, flags, NULL);
-}
-
-/***********************************************************************
- *             CM_Get_Device_ID_List_Size_ExA  (SETUPAPI.@)
- */
-CONFIGRET WINAPI CM_Get_Device_ID_List_Size_ExA(ULONG *len, const char *filter, ULONG flags, HMACHINE machine)
-{
-    WCHAR *wfilter = NULL;
-    unsigned int slen;
+ *             CM_Get_Device_ID_List_SizeA  (SETUPAPI.@)
+ */
+CONFIGRET WINAPI CM_Get_Device_ID_List_SizeA(ULONG *len, const char *filter, ULONG flags)
+{
+    WCHAR *wfilter = NULL, *p;
     CONFIGRET ret;
 
     TRACE("%p %s %#lx.\n", len, debugstr_a(filter), flags);
 
-    if (machine)
-        FIXME("machine %p.\n", machine);
-
     if (filter)
     {
-        slen = strlen(filter) + 1;
-        if (!(wfilter = malloc(slen * sizeof(*wfilter))))
+        if (!(wfilter = malloc(strlen(filter) * sizeof(*wfilter))))
             return CR_OUT_OF_MEMORY;
-        MultiByteToWideChar(CP_ACP, 0, filter, slen, wfilter, slen);
+
+        p = wfilter;
+        while ((*p++ = *filter++))
+            ;
     }
     ret = CM_Get_Device_ID_List_SizeW(len, wfilter, flags);
     free(wfilter);
     return ret;
-}
-
-/***********************************************************************
- *             CM_Get_Device_ID_List_SizeA  (SETUPAPI.@)
- */
-CONFIGRET WINAPI CM_Get_Device_ID_List_SizeA(ULONG *len, const char *filter, ULONG flags)
-{
-    return CM_Get_Device_ID_List_Size_ExA(len, filter, flags, NULL);
 }
 
 static CONFIGRET get_device_interface_list(const GUID *class_guid, DEVINSTID_W device_id, WCHAR *buffer, ULONG *len,
