diff --git a/dlls/windows.media.speech/tests/speech.c b/dlls/windows.media.speech/tests/speech.c
--- a/dlls/windows.media.speech/tests/speech.c
+++ b/dlls/windows.media.speech/tests/speech.c
@@ -202,7 +202,23 @@
                                                   ISpeechContinuousRecognitionSession *sender,
                                                   ISpeechContinuousRecognitionResultGeneratedEventArgs *args )
 {
-    trace("iface %p, sender %p, args %p.\n", iface, sender, args);
+    ISpeechRecognitionResult *result;
+    HSTRING hstring;
+    HRESULT hr;
+
+    if (!args) return S_OK;
+
+    hr = ISpeechContinuousRecognitionResultGeneratedEventArgs_get_Result(args, &result);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = ISpeechRecognitionResult_get_Text(result, &hstring);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    trace("iface %p, sender %p, args %p, text %s.\n", iface, sender, args, debugstr_w(WindowsGetStringRawBuffer(hstring, NULL)));
+
+    WindowsDeleteString(hstring);
+    ISpeechRecognitionResult_Release(result);
+
     return S_OK;
 }
 
@@ -430,9 +446,9 @@
 
     async_id = 0xdeadbeef;
     hr = IAsyncInfo_get_Id(async_info, &async_id);
-    if (expect_status < 4) ok_(__FILE__, line)(hr == S_OK, "IAsyncInfo_get_Id returned %#lx\n", hr);
-    else ok_(__FILE__, line)(hr == E_ILLEGAL_METHOD_CALL, "IAsyncInfo_get_Id returned %#lx\n", hr);
-    todo_wine_if(expect_id != 1) ok_(__FILE__, line)(async_id == expect_id, "got async_id %#x\n", async_id);
+    if (expect_status < 4) todo_wine ok_(__FILE__, line)(hr == S_OK, "IAsyncInfo_get_Id returned %#lx\n", hr);
+    else todo_wine ok_(__FILE__, line)(hr == E_ILLEGAL_METHOD_CALL, "IAsyncInfo_get_Id returned %#lx\n", hr);
+    todo_wine ok_(__FILE__, line)(async_id == expect_id, "got async_id %#x\n", async_id);
 
     async_status = 0xdeadbeef;
     hr = IAsyncInfo_get_Status(async_info, &async_status);
@@ -671,6 +687,66 @@
     impl->iterator = &iterator->IIterator_HSTRING_iface;
 
     return S_OK;
+}
+
+#define check_comparable_presence(a, b) _check_comparable_presence( __LINE__, (a), (b))
+static void _check_comparable_presence( unsigned line, IVectorView_VoiceInformation *voices, IVoiceInformation *voice)
+{
+    HSTRING in_display, in_id, in_language;
+    HSTRING vc_display, vc_id, vc_language;
+    IVoiceInformation *vc_voice;
+    enum VoiceGender in_gender, vc_gender;
+    UINT32 size, idx, found_count = 0;
+    HRESULT hr;
+    INT32 cmp;
+
+    hr = IVoiceInformation_get_DisplayName(voice, &in_display);
+    ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IVoiceInformation_get_Id(voice, &in_id);
+    ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IVoiceInformation_get_Language(voice, &in_language);
+    ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IVoiceInformation_get_Gender(voice, &in_gender);
+    ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    hr = IVectorView_VoiceInformation_get_Size(voices, &size);
+    ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    for (idx = 0; SUCCEEDED(hr = IVectorView_VoiceInformation_GetAt(voices, idx, &vc_voice)); idx++)
+    {
+        hr = IVoiceInformation_get_DisplayName(vc_voice, &vc_display);
+        ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+        hr = IVoiceInformation_get_Id(vc_voice, &vc_id);
+        ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+        hr = IVoiceInformation_get_Language(vc_voice, &vc_language);
+        ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+        hr = IVoiceInformation_get_Gender(vc_voice, &vc_gender);
+        ok_(__FILE__, line)(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+        trace("%u] %s/%s/%s/%u\n",
+              idx - 1, debugstr_hstring(vc_display), debugstr_hstring(vc_id), debugstr_hstring(vc_language), vc_gender);
+
+        if (SUCCEEDED(WindowsCompareStringOrdinal(in_display, vc_display, &cmp)) && !cmp &&
+            SUCCEEDED(WindowsCompareStringOrdinal(in_id, vc_id, &cmp)) && !cmp &&
+            SUCCEEDED(WindowsCompareStringOrdinal(in_language, vc_language, &cmp)) && !cmp &&
+            in_gender == vc_gender)
+        {
+            found_count++;
+        }
+        WindowsDeleteString(vc_display);
+        WindowsDeleteString(vc_id);
+        WindowsDeleteString(vc_language);
+        IVoiceInformation_Release(vc_voice);
+    }
+    ok(hr == E_BOUNDS, "Got unexpected hr %#lx.\n", hr);
+    ok(idx != 0, "Vector view shouldn't be empty!\n");
+    ok(idx == size, "Incoherent index/size %u/%u!\n", idx, size);
+
+    ok_(__FILE__, line)(found_count == 1, "Found several (%u) instances of %s/%s/%s/%u\n",
+                        found_count,
+                        debugstr_hstring(in_display), debugstr_hstring(in_id), debugstr_hstring(in_language), in_gender);
+
+    WindowsDeleteString(in_display);
+    WindowsDeleteString(in_id);
+    WindowsDeleteString(in_language);
 }
 
 static void test_ActivationFactory(void)
@@ -789,7 +865,7 @@
     IVectorView_IMediaMarker *media_markers = NULL;
     IVectorView_VoiceInformation *voices = NULL;
     IInstalledVoicesStatic *voices_static = NULL;
-    ISpeechSynthesisStream *ss_stream = NULL, *tmp;
+    ISpeechSynthesisStream *ss_stream = NULL;
     IRandomAccessStream *ra_stream;
     IInputStream *inp_stream;
     IBuffer *buffer = NULL, *buffer2 = NULL;
@@ -801,11 +877,13 @@
     IClosable *closable;
     struct async_inspectable_handler async_inspectable_handler;
     HMODULE hdll;
-    HSTRING str, str2;
+    HSTRING str, str2, default_voice_id;
     UINT64 value;
     HRESULT hr;
-    UINT32 size;
+    UINT32 size, idx;
+    BOOLEAN found;
     ULONG ref;
+    INT32 cmp;
 
     hr = RoInitialize(RO_INIT_MULTITHREADED);
     ok(hr == S_OK, "RoInitialize failed, hr %#lx\n", hr);
@@ -887,7 +965,7 @@
     size = 0xdeadbeef;
     hr = IVectorView_VoiceInformation_get_Size(voices, &size);
     ok(hr == S_OK, "IVectorView_VoiceInformation_get_Size voices failed, hr %#lx\n", hr);
-    todo_wine ok(size != 0 && size != 0xdeadbeef, "IVectorView_VoiceInformation_get_Size returned %u\n", size);
+    ok(size != 0 && size != 0xdeadbeef, "IVectorView_VoiceInformation_get_Size returned %u\n", size);
 
     voice = (IVoiceInformation *)0xdeadbeef;
     hr = IVectorView_VoiceInformation_GetAt(voices, size, &voice);
@@ -898,20 +976,27 @@
     ok(hr == S_OK, "IVectorView_VoiceInformation_GetMany failed, hr %#lx\n", hr);
     ok(size == 0, "IVectorView_VoiceInformation_GetMany returned count %u\n", size);
 
-    IVectorView_VoiceInformation_Release(voices);
-
     hr = IInstalledVoicesStatic_get_DefaultVoice(voices_static, &voice);
-    todo_wine ok(hr == S_OK, "IInstalledVoicesStatic_get_DefaultVoice failed, hr %#lx\n", hr);
-
-    if (hr == S_OK)
-    {
-        IVoiceInformation_get_Description(voice, &str2);
-        trace("SpeechSynthesizer default voice %s.\n", debugstr_hstring(str2));
-
-        WindowsDeleteString(str2);
-        ref = IVoiceInformation_Release(voice);
-        ok(ref == 0, "Got unexpected ref %lu.\n", ref);
-    }
+    ok(hr == S_OK, "IInstalledVoicesStatic_get_DefaultVoice failed, hr %#lx\n", hr);
+
+    /* check that VoiceInformation in static vector voice are not shared when exposed to user */
+    idx = size;
+    hr = IVectorView_VoiceInformation_IndexOf(voices, voice, &idx, &found);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    ok(!found, "Shouldn't find default element\n");
+
+    check_comparable_presence(voices, voice);
+
+    hr = IVoiceInformation_get_Description(voice, &str2);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+    trace("SpeechSynthesizer default voice %s.\n", debugstr_hstring(str2));
+    WindowsDeleteString(str2);
+
+    hr = IVoiceInformation_get_Id(voice, &default_voice_id);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    ref = IVoiceInformation_Release(voice);
+    ok(ref == 0, "Got unexpected ref %lu.\n", ref);
 
     IInstalledVoicesStatic_Release(voices_static);
     IAgileObject_Release(agile_object);
@@ -926,6 +1011,23 @@
     WindowsDeleteString(str);
 
     hr = IInspectable_QueryInterface(inspectable, &IID_ISpeechSynthesizer, (void **)&synthesizer);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = ISpeechSynthesizer_get_Voice(synthesizer, &voice);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = IVoiceInformation_get_Id(voice, &str);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = WindowsCompareStringOrdinal(str, default_voice_id, &cmp);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    hr = WindowsDeleteString(str);
+    ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+    IVoiceInformation_Release(voice);
+
+    hr = WindowsDeleteString(default_voice_id);
     ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
     /* Test SynthesizeTextToStreamAsync */
@@ -984,12 +1086,6 @@
     ref = IBuffer_Release(buffer2);
     ok(!ref, "got refcount %ld.\n", ref);
 
-    tmp = (void *)0xdeadbeef;
-    hr = IAsyncOperation_SpeechSynthesisStream_GetResults(operation_ss_stream, &tmp);
-    ok(hr == S_OK, "IAsyncOperation_SpeechSynthesisStream_GetResults failed, hr %#lx\n", hr);
-    todo_wine ok(tmp == NULL, "Got %p.\n", tmp);
-    if (tmp && tmp != (void *)0xdeadbeef) ISpeechSynthesisStream_Release(tmp);
-
     hr = ISpeechSynthesisStream_get_Markers(ss_stream, &media_markers);
     ok(hr == S_OK, "ISpeechSynthesisStream_get_Markers failed, hr %#lx\n", hr);
     check_interface(media_markers, &IID_IVectorView_IMediaMarker, TRUE);
@@ -1000,7 +1096,7 @@
     ok(ref == 0, "Got unexpected ref %lu.\n", ref);
 
     ref = ISpeechSynthesisStream_Release(ss_stream);
-    todo_wine ok(ref == 0, "Got unexpected ref %lu.\n", ref);
+    ok(ref == 0, "Got unexpected ref %lu.\n", ref);
 
     IAsyncOperation_SpeechSynthesisStream_Release(operation_ss_stream);
 
@@ -1024,7 +1120,7 @@
     check_interface(ss_stream, &IID_IAgileObject, TRUE);
 
     ref = ISpeechSynthesisStream_Release(ss_stream);
-    todo_wine ok(ref == 0, "Got unexpected ref %lu.\n", ref);
+    ok(ref == 0, "Got unexpected ref %lu.\n", ref);
 
     IAsyncOperation_SpeechSynthesisStream_Release(operation_ss_stream);
 
@@ -1067,20 +1163,58 @@
         ISpeechSynthesizerOptions *options;
 
         hr = ISpeechSynthesizer2_get_Options(synthesizer2, &options);
-        todo_wine ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+        ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
 
         if (hr == S_OK)
         {
+            ISpeechSynthesizerOptions2 *options2;
             ISpeechSynthesizerOptions3 *options3;
+            boolean bool_value;
+            DOUBLE double_value;
+            enum SpeechAppendedSilence silence_value;
+            enum SpeechPunctuationSilence punctuation_value;
 
             check_interface(options, &IID_IAgileObject, TRUE);
+            bool_value = 0xff;
+            hr = ISpeechSynthesizerOptions_get_IncludeSentenceBoundaryMetadata(options, &bool_value);
+            ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+            ok(! bool_value, "Got unepected option %u\n",  bool_value);
+            bool_value = 0xff;
+            hr = ISpeechSynthesizerOptions_get_IncludeWordBoundaryMetadata(options, &bool_value);
+            ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+            ok(!bool_value, "Got unepected option %u\n", bool_value);
+
             check_optional_interface(options, &IID_ISpeechSynthesizerOptions2, TRUE); /* Requires Win10 >= 1709 */
+            hr = ISpeechSynthesizerOptions_QueryInterface(options, &IID_ISpeechSynthesizerOptions2, (void **)&options2);
+            ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+
+            hr = ISpeechSynthesizerOptions2_get_AudioPitch(options2, &double_value);
+            ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+            ok(double_value == 1.0f, "Got unepected option %f\n", double_value);
+
+            hr = ISpeechSynthesizerOptions2_get_AudioVolume(options2, &double_value);
+            ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+            ok(double_value == 1.0f, "Got unepected option %f\n", double_value);
+
+            hr = ISpeechSynthesizerOptions2_get_SpeakingRate(options2, &double_value);
+            ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+            ok(double_value == 1.0f, "Got unepected option %f\n", double_value);
+
+            ISpeechSynthesizerOptions2_Release(options2);
 
             hr = ISpeechSynthesizerOptions_QueryInterface(options, &IID_ISpeechSynthesizerOptions3, (void **)&options3);
             ok(hr == S_OK || broken(hr == E_NOINTERFACE), "Got unexpected hr %#lx.\n", hr); /* Requires Win10 >= 1803 */
 
             if (hr == S_OK)
             {
+                hr = ISpeechSynthesizerOptions3_get_AppendedSilence(options3, &silence_value);
+                ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+                ok(silence_value == SpeechAppendedSilence_Default, "Got unepected option %u\n", silence_value);
+
+                hr = ISpeechSynthesizerOptions3_get_PunctuationSilence(options3, &punctuation_value);
+                ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
+                ok(punctuation_value == SpeechPunctuationSilence_Default, "Got unepected option %u\n", punctuation_value);
+
                 ref = ISpeechSynthesizerOptions3_Release(options3);
                 ok(ref == 2, "Got unexpected ref %lu.\n", ref);
             }
@@ -1231,7 +1365,7 @@
     ok(ref == 1, "Got unexpected ref %lu.\n", ref);
 
     hr = RoActivateInstance(hstr, &inspectable);
-    ok(hr == S_OK || broken(hr == SPERR_WINRT_INTERNAL_ERROR), "Got unexpected hr %#lx.\n", hr);
+    ok(hr == S_OK || hr == SPERR_WINRT_INTERNAL_ERROR, "Got unexpected hr %#lx.\n", hr);
 
     if (hr == S_OK)
     {
@@ -1330,13 +1464,10 @@
         ok(result_status == SpeechRecognitionResultStatus_Success, "Got unexpected status %#x.\n", result_status);
 
         ref = ISpeechRecognitionCompilationResult_Release(compilation_result);
-        todo_wine ok(!ref , "Got unexpected ref %lu.\n", ref);
-
-        compilation_result = (void *)0xdeadbeef;
+        ok(!ref , "Got unexpected ref %lu.\n", ref);
+
         hr = IAsyncOperation_SpeechRecognitionCompilationResult_GetResults(operation, &compilation_result);
-        todo_wine ok(hr == E_UNEXPECTED, "Got unexpected hr %#lx.\n", hr);
-        todo_wine ok(compilation_result == NULL, "Got %p.\n", compilation_result);
-        if (compilation_result && compilation_result != (void *)0xdeadbeef) ISpeechRecognitionCompilationResult_Release(compilation_result);
+        ok(hr == E_UNEXPECTED, "Got unexpected hr %#lx.\n", hr);
 
         hr = IAsyncOperation_SpeechRecognitionCompilationResult_QueryInterface(operation, &IID_IAsyncInfo, (void **)&info);
         ok(hr == S_OK, "Got unexpected hr %#lx.\n", hr);
@@ -1453,7 +1584,7 @@
     }
     else if (hr == SPERR_WINRT_INTERNAL_ERROR) /* Not sure when this triggers. Probably if a language pack is not installed. */
     {
-        win_skip("Could not init SpeechRecognizer with default language!\n");
+        skip("Could not init SpeechRecognizer with default language!\n");
     }
 
 done:
@@ -1629,7 +1760,7 @@
     static const WCHAR *list_constraint_name = L"Windows.Media.SpeechRecognition.SpeechRecognitionListConstraint";
     static const WCHAR *recognizer_name = L"Windows.Media.SpeechRecognition.SpeechRecognizer";
     static const WCHAR *speech_constraint_tag = L"test_message";
-    static const WCHAR *speech_constraints[] = { L"This is a test.", L"Number 5!", L"What time is it?" };
+    static const WCHAR *speech_constraints[] = { L"This is a test", L"Number 5", L"What time is it" };
     ISpeechRecognitionListConstraintFactory *listconstraint_factory = NULL;
     IAsyncOperation_SpeechRecognitionCompilationResult *operation = NULL;
     IVector_ISpeechRecognitionConstraint *constraints = NULL;
@@ -1670,12 +1801,12 @@
     ok(hr == S_OK, "WindowsCreateString failed, hr %#lx.\n", hr);
 
     hr = RoActivateInstance(hstr, &inspectable);
-    ok(hr == S_OK || broken(hr == SPERR_WINRT_INTERNAL_ERROR || hr == REGDB_E_CLASSNOTREG), "Got unexpected hr %#lx.\n", hr);
+    ok(hr == S_OK || hr == SPERR_WINRT_INTERNAL_ERROR || broken(hr == REGDB_E_CLASSNOTREG), "Got unexpected hr %#lx.\n", hr);
     WindowsDeleteString(hstr);
 
-    if (FAILED(hr))  /* Win 8 and 8.1 and Win10 without enabled SR. */
-    {
-        win_skip("SpeechRecognizer cannot be activated!\n");
+    if (FAILED(hr))  /* Win 8 and 8.1 and Win10 without enabled SR. Wine with missing Unix side dependencies. */
+    {
+        skip("SpeechRecognizer cannot be activated!\n");
         goto done;
     }
 
@@ -1792,6 +1923,8 @@
     ok(hr == S_OK, "ISpeechRecognizer2_get_State failed, hr %#lx.\n", hr);
     ok(recog_state == SpeechRecognizerState_Capturing || broken(recog_state == SpeechRecognizerState_Idle), "recog_state was %u.\n", recog_state);
 
+
+    Sleep(10000);
     /*
      * TODO: Use a loopback device together with prerecorded audio files to test the recognizer's functionality.
      */
