diff --git a/dlls/wined3d/swapchain.c b/dlls/wined3d/swapchain.c
--- a/dlls/wined3d/swapchain.c
+++ b/dlls/wined3d/swapchain.c
@@ -27,19 +27,6 @@
 WINE_DEFAULT_DEBUG_CHANNEL(d3d);
 WINE_DECLARE_DEBUG_CHANNEL(d3d_perf);
 
-static BOOL set_window_present_rect(HWND hwnd, UINT x, UINT y, UINT width, UINT height)
-{
-    RECT rect = {x, y, x + width, y + height};
-    D3DKMT_ESCAPE escape = {0};
-
-    escape.Type = D3DKMT_ESCAPE_SET_PRESENT_RECT_WINE;
-    escape.hContext = HandleToULong(hwnd);
-    escape.pPrivateDriverData = &rect;
-    escape.PrivateDriverDataSize = sizeof(rect);
-
-    return !D3DKMTEscape(&escape);
-}
-
 void wined3d_swapchain_cleanup(struct wined3d_swapchain *swapchain)
 {
     HRESULT hr;
@@ -117,7 +104,7 @@
 
     vk_info = &wined3d_adapter_vk(device_vk->d.adapter)->vk_info;
 
-    if ((vr = VK_CALL(vkQueueWaitIdle(device_vk->graphics_queue.vk_queue))) < 0)
+    if ((vr = VK_CALL(vkQueueWaitIdle(device_vk->vk_queue))) < 0)
         ERR("Failed to wait on queue, vr %s.\n", wined3d_debug_vkresult(vr));
     free(swapchain_vk->vk_images);
     for (i = 0; i < swapchain_vk->image_count; ++i)
@@ -171,9 +158,6 @@
             wined3d_device_uninit_3d(device);
         wined3d_cs_finish(device->cs, WINED3D_CS_QUEUE_DEFAULT);
 
-        if (swapchain->dc)
-            wined3d_release_dc(swapchain->win_handle, swapchain->dc);
-
         swapchain->parent_ops->wined3d_object_destroyed(swapchain->parent);
         swapchain->device->adapter->adapter_ops->adapter_destroy_swapchain(swapchain);
 
@@ -202,20 +186,13 @@
 
     wined3d_cs_finish(swapchain->device->cs, WINED3D_CS_QUEUE_DEFAULT);
 
-    if (swapchain->dc)
-        wined3d_release_dc(swapchain->win_handle, swapchain->dc);
-
     swapchain->win_handle = window;
-
-    if (!(swapchain->dc = GetDCEx(swapchain->win_handle, 0, DCX_USESTYLE | DCX_CACHE)))
-        WARN("Failed to retrieve device context, trying swapchain backup.\n");
 }
 
 HRESULT CDECL wined3d_swapchain_present(struct wined3d_swapchain *swapchain,
         const RECT *src_rect, const RECT *dst_rect, HWND dst_window_override,
         unsigned int swap_interval, uint32_t flags)
 {
-    const struct wined3d_swapchain_desc *desc = &swapchain->state.desc;
     RECT s, d;
 
     TRACE("swapchain %p, src_rect %s, dst_rect %s, dst_window_override %p, swap_interval %u, flags %#x.\n",
@@ -236,16 +213,14 @@
 
     if (!src_rect)
     {
-        SetRect(&s, 0, 0, desc->backbuffer_width, desc->backbuffer_height);
+        SetRect(&s, 0, 0, swapchain->state.desc.backbuffer_width,
+                swapchain->state.desc.backbuffer_height);
         src_rect = &s;
     }
 
     if (!dst_rect)
     {
-        if (!desc->windowed)
-            SetRect(&d, 0, 0, desc->backbuffer_width, desc->backbuffer_height);
-        else
-            GetClientRect(swapchain->win_handle, &d);
+        GetClientRect(swapchain->win_handle, &d);
         dst_rect = &d;
     }
 
@@ -429,9 +404,9 @@
     D3DKMT_CREATEDCFROMMEMORY create_desc;
     const struct wined3d_format *format;
     unsigned int row_pitch, slice_pitch;
+    HDC src_dc, dst_dc;
     NTSTATUS status;
     HBITMAP bitmap;
-    HDC src_dc;
 
     static unsigned int once;
 
@@ -472,11 +447,15 @@
 
     TRACE("Created source DC %p, bitmap %p for backbuffer %p.\n", src_dc, bitmap, back_buffer);
 
-    if (!StretchBlt(swapchain->dc, dst_rect->left, dst_rect->top, dst_rect->right - dst_rect->left,
+    if (!(dst_dc = GetDCEx(swapchain->win_handle, 0, DCX_USESTYLE | DCX_CACHE)))
+        ERR("Failed to get destination DC.\n");
+
+    if (!StretchBlt(dst_dc, dst_rect->left, dst_rect->top, dst_rect->right - dst_rect->left,
             dst_rect->bottom - dst_rect->top, src_dc, src_rect->left, src_rect->top,
             src_rect->right - src_rect->left, src_rect->bottom - src_rect->top, SRCCOPY))
         ERR("Failed to blit.\n");
 
+    ReleaseDC(swapchain->win_handle, dst_dc);
     destroy_desc.hDc = src_dc;
     destroy_desc.hBitmap = bitmap;
     if ((status = D3DKMTDestroyDCFromMemory(&destroy_desc)))
@@ -583,28 +562,19 @@
 static bool swapchain_present_is_partial_copy(struct wined3d_swapchain *swapchain, const RECT *dst_rect)
 {
     enum wined3d_swap_effect swap_effect = swapchain->state.desc.swap_effect;
-    const struct wined3d_swapchain_desc *desc = &swapchain->state.desc;
-    unsigned int width, height;
+    RECT client_rect;
+    unsigned int t;
 
     if (swap_effect != WINED3D_SWAP_EFFECT_COPY && swap_effect != WINED3D_SWAP_EFFECT_COPY_VSYNC)
         return false;
 
-    if (!desc->windowed)
-    {
-        width = desc->backbuffer_width;
-        height = desc->backbuffer_height;
-    }
-    else
-    {
-        RECT client_rect;
-        GetClientRect(swapchain->win_handle, &client_rect);
-        width = client_rect.right - client_rect.left;
-        height = client_rect.bottom - client_rect.top;
-    }
-
-    if ((dst_rect->left && dst_rect->right) || abs(dst_rect->right - dst_rect->left) != width)
+    GetClientRect(swapchain->win_handle, &client_rect);
+
+    t = client_rect.right - client_rect.left;
+    if ((dst_rect->left && dst_rect->right) || abs(dst_rect->right - dst_rect->left) != t)
         return true;
-    if ((dst_rect->top && dst_rect->bottom) || abs(dst_rect->bottom - dst_rect->top) != height)
+    t = client_rect.bottom - client_rect.top;
+    if ((dst_rect->top && dst_rect->bottom) || abs(dst_rect->bottom - dst_rect->top) != t)
         return true;
 
     return false;
@@ -928,7 +898,7 @@
     swapchain_vk->vk_surface = vk_surface;
 
     if ((vr = VK_CALL(vkGetPhysicalDeviceSurfaceSupportKHR(adapter_vk->physical_device,
-            device_vk->graphics_queue.vk_queue_family_index, vk_surface, &supported))) < 0 || !supported)
+            device_vk->vk_queue_family_index, vk_surface, &supported))) < 0 || !supported)
     {
         ERR("Queue family does not support presentation on this surface, vr %s.\n", wined3d_debug_vkresult(vr));
         goto fail;
@@ -1037,7 +1007,6 @@
 
 fail:
     VK_CALL(vkDestroySurfaceKHR(vk_info->instance, vk_surface, NULL));
-    swapchain_vk->vk_surface = 0;
     return E_FAIL;
 }
 
@@ -1190,7 +1159,7 @@
     present_desc.pSwapchains = &swapchain_vk->vk_swapchain;
     present_desc.pImageIndices = &image_idx;
     present_desc.pResults = NULL;
-    if ((vr = VK_CALL(vkQueuePresentKHR(device_vk->graphics_queue.vk_queue, &present_desc))))
+    if ((vr = VK_CALL(vkQueuePresentKHR(device_vk->vk_queue, &present_desc))))
         WARN("Present returned vr %s.\n", wined3d_debug_vkresult(vr));
     return vr;
 }
@@ -1306,9 +1275,9 @@
 {
     struct wined3d_dc_info *front;
     POINT offset = {0, 0};
+    HDC src_dc, dst_dc;
     RECT draw_rect;
     HWND window;
-    HDC src_dc;
 
     TRACE("swapchain %p.\n", swapchain);
 
@@ -1323,6 +1292,7 @@
 
     src_dc = front->dc;
     window = swapchain->win_handle;
+    dst_dc = GetDCEx(window, 0, DCX_CLIPSIBLINGS | DCX_CACHE);
 
     /* Front buffer coordinates are screen coordinates. Map them to the
      * destination window if not fullscreened. */
@@ -1335,9 +1305,10 @@
             swapchain->front_buffer->resource.height);
     IntersectRect(&draw_rect, &draw_rect, &swapchain->front_buffer_update);
 
-    BitBlt(swapchain->dc, draw_rect.left - offset.x, draw_rect.top - offset.y,
+    BitBlt(dst_dc, draw_rect.left - offset.x, draw_rect.top - offset.y,
             draw_rect.right - draw_rect.left, draw_rect.bottom - draw_rect.top,
             src_dc, draw_rect.left, draw_rect.top, SRCCOPY);
+    ReleaseDC(window, dst_dc);
 
     SetRectEmpty(&swapchain->front_buffer_update);
 }
@@ -1347,8 +1318,7 @@
 {
     struct wined3d_dc_info *front, *back;
     HBITMAP bitmap;
-    void *heap_pointer;
-    void *heap_memory;
+    void *data;
     HDC dc;
 
     front = &swapchain->front_buffer->dc_info[0];
@@ -1357,18 +1327,15 @@
     /* Flip the surface data. */
     dc = front->dc;
     bitmap = front->bitmap;
-    heap_pointer = swapchain->front_buffer->resource.heap_pointer;
-    heap_memory = swapchain->front_buffer->resource.heap_memory;
+    data = swapchain->front_buffer->resource.heap_memory;
 
     front->dc = back->dc;
     front->bitmap = back->bitmap;
-    swapchain->front_buffer->resource.heap_pointer = swapchain->back_buffers[0]->resource.heap_pointer;
     swapchain->front_buffer->resource.heap_memory = swapchain->back_buffers[0]->resource.heap_memory;
 
     back->dc = dc;
     back->bitmap = bitmap;
-    swapchain->back_buffers[0]->resource.heap_pointer = heap_pointer;
-    swapchain->back_buffers[0]->resource.heap_memory = heap_memory;
+    swapchain->back_buffers[0]->resource.heap_memory = data;
 
     SetRect(&swapchain->front_buffer_update, 0, 0,
             swapchain->front_buffer->resource.width,
@@ -1579,9 +1546,6 @@
     swapchain->swap_interval = WINED3D_SWAP_INTERVAL_DEFAULT;
     swapchain_set_max_frame_latency(swapchain, device);
 
-    if (!(swapchain->dc = GetDCEx(swapchain->win_handle, 0, DCX_USESTYLE | DCX_CACHE)))
-        WARN("Failed to retrieve device context, trying swapchain backup.\n");
-
     if (!swapchain->state.desc.windowed)
     {
         if (FAILED(hr = wined3d_output_get_desc(desc->output, &output_desc)))
@@ -1712,9 +1676,6 @@
         wined3d_texture_decref(swapchain->front_buffer);
     }
 
-    if (swapchain->dc)
-        wined3d_release_dc(swapchain->win_handle, swapchain->dc);
-
     wined3d_swapchain_state_cleanup(&swapchain->state);
     wined3d_mutex_unlock();
 
@@ -1763,10 +1724,7 @@
     }
 
     if (FAILED(hr = wined3d_swapchain_vk_create_vulkan_swapchain(swapchain_vk)))
-    {
         WARN("Failed to create a Vulkan swapchain, hr %#lx.\n", hr);
-        return hr;
-    }
 
     return WINED3D_OK;
 }
@@ -2135,22 +2093,6 @@
     }
     else
     {
-        if (FAILED(hr = wined3d_output_get_desc(state->desc.output, &output_desc)))
-        {
-            ERR("Failed to get output description, hr %#lx.\n", hr);
-            wined3d_mutex_unlock();
-            return hr;
-        }
-        width = output_desc.desktop_rect.right - output_desc.desktop_rect.left;
-        height = output_desc.desktop_rect.bottom - output_desc.desktop_rect.top;
-
-        GetWindowRect(window, &window_rect);
-        if (width != window_rect.right - window_rect.left || height != window_rect.bottom - window_rect.top)
-        {
-            TRACE("Update saved window state.\n");
-            state->original_window_rect = window_rect;
-        }
-
         if (state->desc.flags & WINED3D_SWAPCHAIN_ALLOW_MODE_SWITCH)
         {
             actual_mode = *mode;
@@ -2161,18 +2103,19 @@
                 wined3d_mutex_unlock();
                 return hr;
             }
-            if (FAILED(hr = wined3d_output_get_desc(state->desc.output, &output_desc)))
-            {
-                ERR("Failed to get output description, hr %#lx.\n", hr);
-                wined3d_mutex_unlock();
-                return hr;
-            }
-
-            width = output_desc.desktop_rect.right - output_desc.desktop_rect.left;
-            height = output_desc.desktop_rect.bottom - output_desc.desktop_rect.top;
-        }
+        }
+
+        if (FAILED(hr = wined3d_output_get_desc(state->desc.output, &output_desc)))
+        {
+            ERR("Failed to get output description, hr %#lx.\n", hr);
+            wined3d_mutex_unlock();
+            return hr;
+        }
+
         x = output_desc.desktop_rect.left;
         y = output_desc.desktop_rect.top;
+        width = output_desc.desktop_rect.right - output_desc.desktop_rect.left;
+        height = output_desc.desktop_rect.bottom - output_desc.desktop_rect.top;
     }
 
     wined3d_mutex_unlock();
@@ -2247,7 +2190,6 @@
     static const UINT timeout = 1500;
     DWORD window_tid = GetWindowThreadProcessId(s->window, NULL);
     DWORD tid = GetCurrentThreadId();
-    HANDLE thread;
 
     TRACE("Window %p belongs to thread %#lx.\n", s->window, window_tid);
     /* If the window belongs to a different thread, modifying the style and/or
@@ -2265,18 +2207,8 @@
             else
                 KillTimer(s->window, WINED3D_WINDOW_TOPMOST_TIMER_ID);
         }
-
-        set_window_state_thread(s);
-    }
-    else if ((thread = CreateThread(NULL, 0, set_window_state_thread, s, 0, NULL)))
-    {
-        SetThreadDescription(thread, L"wined3d_set_window_state");
-        CloseHandle(thread);
-    }
-    else
-    {
-        ERR("Failed to create thread.\n");
-    }
+    }
+    set_window_state_thread(s);
 }
 
 HRESULT wined3d_swapchain_state_setup_fullscreen(struct wined3d_swapchain_state *state,
@@ -2291,8 +2223,6 @@
         WARN("%p is not a valid window.\n", window);
         return WINED3DERR_NOTAVAILABLE;
     }
-
-    set_window_present_rect(window, x, y, width, height);
 
     if (!(s = malloc(sizeof(*s))))
         return E_OUTOFMEMORY;
@@ -2336,8 +2266,6 @@
 {
     struct wined3d_window_state *s;
     LONG style, exstyle;
-
-    set_window_present_rect(window, 0, 0, 0, 0);
 
     if (!state->style && !state->exstyle)
         return;
@@ -2482,9 +2410,6 @@
             HWND window = state->device_window;
             BOOL filter;
 
-            set_window_present_rect(state->device_window, output_desc.desktop_rect.left,
-                    output_desc.desktop_rect.top, width, height);
-
             /* Fullscreen -> fullscreen mode change */
             filter = wined3d_filter_messages(window, TRUE);
             MoveWindow(window, output_desc.desktop_rect.left, output_desc.desktop_rect.top, width,
