diff --git a/dlls/ntdll/unix/unix_private.h b/dlls/ntdll/unix/unix_private.h
--- a/dlls/ntdll/unix/unix_private.h
+++ b/dlls/ntdll/unix/unix_private.h
@@ -101,42 +101,26 @@
 /* thread private data, stored in NtCurrentTeb()->GdiTebBatch */
 struct ntdll_thread_data
 {
-    void                     *cpu_data[16];  /* 1d4/02f0 reserved for CPU-specific data */
-    SYSTEM_SERVICE_TABLE     *syscall_table; /* 214/0370 syscall table */
-    struct syscall_frame     *syscall_frame; /* 218/0378 current syscall frame */
-    int                       syscall_trace; /* 21c/0380 syscall trace flag */
-    int                       request_fd;    /* fd for sending server requests */
-    int                       reply_fd;      /* fd for receiving server replies */
-    int                       wait_fd[2];    /* fd for sleeping server requests */
-    int                       alert_fd;      /* inproc sync fd for user apc alerts */
-    BOOL                      allow_writes;  /* ThreadAllowWrites flags */
-    pthread_t                 pthread_id;    /* pthread thread id */
-    void                     *kernel_stack;  /* stack for thread startup and kernel syscalls */
-    struct list               entry;         /* entry in TEB list */
-    PRTL_THREAD_START_ROUTINE start;         /* thread entry point */
-    void                     *param;         /* thread entry point parameter */
-    void                     *jmp_buf;       /* setjmp buffer for exception handling */
+    void              *cpu_data[16];  /* reserved for CPU-specific data */
+    void              *kernel_stack;  /* stack for thread startup and kernel syscalls */
+    int                esync_apc_fd;  /* fd to wait on for user APCs */
+    int               *fsync_apc_futex;
+    int                request_fd;    /* fd for sending server requests */
+    int                reply_fd;      /* fd for receiving server replies */
+    int                wait_fd[2];    /* fd for sleeping server requests */
+    BOOL               allow_writes;  /* ThreadAllowWrites flags */
+    pthread_t          pthread_id;    /* pthread thread id */
+    struct list        entry;         /* entry in TEB list */
+    PRTL_THREAD_START_ROUTINE start;  /* thread entry point */
+    void              *param;         /* thread entry point parameter */
+    void              *jmp_buf;       /* setjmp buffer for exception handling */
 };
 
 C_ASSERT( sizeof(struct ntdll_thread_data) <= sizeof(((TEB *)0)->GdiTebBatch) );
-#ifdef _WIN64
-C_ASSERT( offsetof( TEB, GdiTebBatch ) + offsetof( struct ntdll_thread_data, syscall_table ) == 0x370 );
-C_ASSERT( offsetof( TEB, GdiTebBatch ) + offsetof( struct ntdll_thread_data, syscall_frame ) == 0x378 );
-C_ASSERT( offsetof( TEB, GdiTebBatch ) + offsetof( struct ntdll_thread_data, syscall_trace ) == 0x380 );
-#else
-C_ASSERT( offsetof( TEB, GdiTebBatch ) + offsetof( struct ntdll_thread_data, syscall_table ) == 0x214 );
-C_ASSERT( offsetof( TEB, GdiTebBatch ) + offsetof( struct ntdll_thread_data, syscall_frame ) == 0x218 );
-C_ASSERT( offsetof( TEB, GdiTebBatch ) + offsetof( struct ntdll_thread_data, syscall_trace ) == 0x21c );
-#endif
 
 static inline struct ntdll_thread_data *ntdll_get_thread_data(void)
 {
     return (struct ntdll_thread_data *)&NtCurrentTeb()->GdiTebBatch;
-}
-
-static inline struct syscall_frame *get_syscall_frame(void)
-{
-    return ntdll_get_thread_data()->syscall_frame;
 }
 
 /* returns TRUE if the async is complete; FALSE if it should be restarted */
@@ -153,8 +137,9 @@
 static const SIZE_T teb_size = 0x3800;  /* TEB64 + TEB32 + debug info */
 static const SIZE_T signal_stack_mask = 0xffff;
 static const SIZE_T signal_stack_size = 0x10000 - 0x3800;
-static const SIZE_T kernel_stack_size = 0x100000;
-static const SIZE_T min_kernel_stack  = 0x2000;
+extern SIZE_T kernel_stack_size;
+static const SIZE_T kernel_stack_guard_size = 0x1000;
+static const SIZE_T min_kernel_stack  = 0x3000;
 static const LONG teb_offset = 0x2000;
 
 #define FILE_WRITE_TO_END_OF_FILE      ((LONGLONG)-1)
@@ -180,7 +165,6 @@
 extern const char *data_dir;
 extern const char *build_dir;
 extern const char *config_dir;
-extern const char *wineloader;
 extern const char *user_name;
 extern const char **dll_paths;
 extern const char **system_dll_paths;
@@ -197,26 +181,44 @@
 extern unsigned int supported_machines_count;
 extern USHORT supported_machines[8];
 extern BOOL process_exiting;
+extern BOOL terminate_process_running;
+extern LONG terminate_process_exit_code;
 extern HANDLE keyed_event;
-extern int inproc_device_fd;
 extern timeout_t server_start_time;
 extern sigset_t server_block_set;
-extern pthread_mutex_t fd_cache_mutex;
 extern struct _KUSER_SHARED_DATA *user_shared_data;
+extern ULONG process_cookie;
+extern SYSTEM_CPU_INFORMATION cpu_info;
+#ifdef __i386__
+extern struct ldt_copy __wine_ldt_copy;
+#endif
+
+extern BOOL ac_odyssey;
+extern BOOL fsync_simulate_sched_quantum;
+extern BOOL alert_simulate_sched_quantum;
+extern BOOL fsync_yield_to_waiters;
+extern BOOL fsync_help_simulated_pulse;
+extern BOOL localsystem_sid;
+extern BOOL simulate_writecopy;
+extern long long ram_reporting_bias;
+extern BOOL wine_allocs_2g_limit;
+extern char *release_reserved_memory_low_bound;
 
 extern void init_environment(void);
 extern void init_startup_info(void);
 extern void *create_startup_info( const UNICODE_STRING *nt_image, ULONG process_flags,
                                   const RTL_USER_PROCESS_PARAMETERS *params,
                                   const struct pe_image_info *pe_info, DWORD *info_size );
+extern char **build_envp( const WCHAR *envW );
 extern char *get_alternate_wineloader( WORD machine );
 extern NTSTATUS exec_wineloader( char **argv, int socketfd, const struct pe_image_info *pe_info );
-extern NTSTATUS load_builtin( const struct pe_image_info *image_info, UNICODE_STRING *nt_name,
-                              ANSI_STRING *exp_name, USHORT machine, SECTION_IMAGE_INFORMATION *info,
-                              void **module, SIZE_T *size, ULONG_PTR limit_low, ULONG_PTR limit_high, off_t offset );
+extern NTSTATUS load_builtin( const struct pe_image_info *image_info, WCHAR *filename, USHORT machine,
+                              SECTION_IMAGE_INFORMATION *info, void **module, SIZE_T *size,
+                              ULONG_PTR limit_low, ULONG_PTR limit_high, off_t offset  );
 extern BOOL is_builtin_path( const UNICODE_STRING *path, WORD *machine );
-extern NTSTATUS load_main_exe( UNICODE_STRING *nt_name, USHORT load_machine, void **module );
-extern NTSTATUS load_start_exe( UNICODE_STRING *nt_name, void **module );
+extern NTSTATUS load_main_exe( const WCHAR *name, const char *unix_name, const WCHAR *curdir,
+                               USHORT load_machine, WCHAR **image, void **module );
+extern NTSTATUS load_start_exe( WCHAR **image, void **module );
 extern ULONG_PTR redirect_arm64ec_rva( void *module, ULONG_PTR rva, const IMAGE_ARM64EC_METADATA *metadata );
 extern void start_server( BOOL debug );
 
@@ -227,12 +229,12 @@
                                    timeout_t abs_timeout, struct context_data *context, struct user_apc *user_apc );
 extern unsigned int server_wait( const union select_op *select_op, data_size_t size, UINT flags,
                                  const LARGE_INTEGER *timeout );
-extern unsigned int server_wait_for_object( HANDLE handle, BOOL alertable, const LARGE_INTEGER *timeout );
+extern NTSTATUS wait_internal_server( HANDLE handle, BOOLEAN alertable, const LARGE_INTEGER *timeout );
 extern unsigned int server_queue_process_apc( HANDLE process, const union apc_call *call,
                                               union apc_result *result );
 extern int server_get_unix_fd( HANDLE handle, unsigned int wanted_access, int *unix_fd,
                                int *needs_close, enum server_fd_type *type, unsigned int *options );
-extern int wine_server_receive_fd( obj_handle_t *handle );
+extern void wine_server_send_fd( int fd );
 extern void process_exit_wrapper( int status ) DECLSPEC_NORETURN;
 extern size_t server_init_process(void);
 extern void server_init_process_done(void);
@@ -253,8 +255,16 @@
                      | CONTEXT_EXCEPTION_REPORTING | reporting_flag;
 }
 
+extern BOOL xstate_compaction_enabled;
+extern UINT64 xstate_supported_features_mask;
+extern UINT64 xstate_features_size;
 extern unsigned int xstate_get_size( UINT64 compaction_mask, UINT64 mask );
 extern void copy_xstate( XSAVE_AREA_HEADER *dst, XSAVE_AREA_HEADER *src, UINT64 mask );
+
+static inline UINT64 xstate_extended_features(void)
+{
+    return xstate_supported_features_mask & ~(UINT64)3;
+}
 
 extern void set_process_instrumentation_callback( void *callback );
 
@@ -266,12 +276,15 @@
 extern void DECLSPEC_NORETURN exit_process( int status );
 extern void wait_suspend( CONTEXT *context );
 extern NTSTATUS send_debug_event( EXCEPTION_RECORD *rec, CONTEXT *context, BOOL first_chance, BOOL exception );
+extern BOOL validate_context_xstate( CONTEXT *context );
 extern NTSTATUS set_thread_context( HANDLE handle, const void *context, BOOL *self, USHORT machine );
 extern NTSTATUS get_thread_context( HANDLE handle, void *context, BOOL *self, USHORT machine );
 extern unsigned int alloc_object_attributes( const OBJECT_ATTRIBUTES *attr, struct object_attributes **ret,
                                              data_size_t *ret_len );
 extern NTSTATUS system_time_precise( void *args );
-
+extern void get_random( void *buf, ULONG len );
+
+extern void *steamclient_handle_fault( LPCVOID addr, DWORD err );
 extern void *anon_mmap_fixed( void *start, size_t size, int prot, int flags );
 extern void *anon_mmap_alloc( size_t size, int prot );
 extern void virtual_init(void);
@@ -290,10 +303,10 @@
 extern NTSTATUS virtual_alloc_teb( TEB **ret_teb );
 extern void virtual_free_teb( TEB *teb );
 extern NTSTATUS virtual_clear_tls_index( ULONG index );
+extern NTSTATUS virtual_set_tls_information( PROCESS_TLS_INFORMATION *t );
 extern NTSTATUS virtual_alloc_thread_stack( INITIAL_TEB *stack, ULONG_PTR limit_low, ULONG_PTR limit_high,
                                             SIZE_T reserve_size, SIZE_T commit_size, BOOL guard_page );
 extern void virtual_map_user_shared_data(void);
-extern void virtual_init_user_shared_data(void);
 extern NTSTATUS virtual_handle_fault( EXCEPTION_RECORD *rec, void *stack );
 extern unsigned int virtual_locked_server_call( void *req_ptr );
 extern ssize_t virtual_locked_read( int fd, void *addr, size_t size );
@@ -312,21 +325,23 @@
                                             SECTION_IMAGE_INFORMATION *info );
 extern void *get_builtin_so_handle( void *module );
 extern NTSTATUS load_builtin_unixlib( void *module, const char *name );
-
-extern NTSTATUS get_thread_ldt_entry( HANDLE handle, THREAD_DESCRIPTOR_INFORMATION *info, ULONG len );
+extern NTSTATUS unwind_builtin_dll( void *args );
+
+extern NTSTATUS get_thread_ldt_entry( HANDLE handle, void *data, ULONG len, ULONG *ret_len );
 extern void *get_native_context( CONTEXT *context );
 extern void *get_wow_context( CONTEXT *context );
 extern BOOL get_thread_times( int unix_pid, int unix_tid, LARGE_INTEGER *kernel_time,
                               LARGE_INTEGER *user_time );
 extern void signal_init_threading(void);
 extern NTSTATUS signal_alloc_thread( TEB *teb );
+extern void set_thread_teb( TEB *teb );
 extern void signal_free_thread( TEB *teb );
 extern void signal_init_process(void);
 extern void DECLSPEC_NORETURN signal_start_thread( PRTL_THREAD_START_ROUTINE entry, void *arg,
                                                    BOOL suspend, TEB *teb );
 extern SYSTEM_SERVICE_TABLE KeServiceDescriptorTable[4];
 extern void __wine_syscall_dispatcher(void);
-extern void __wine_syscall_dispatcher_return(void);
+extern void DECLSPEC_NORETURN __wine_syscall_dispatcher_return( void *frame, ULONG_PTR retval );
 extern void __wine_unix_call_dispatcher(void);
 extern NTSTATUS signal_set_full_context( CONTEXT *context );
 extern NTSTATUS get_thread_wow64_context( HANDLE handle, void *ctx, ULONG size );
@@ -356,22 +371,23 @@
 extern struct async_fileio *alloc_fileio( DWORD size, async_callback_t callback, HANDLE handle );
 extern void release_fileio( struct async_fileio *io );
 extern NTSTATUS errno_to_status( int err );
+extern BOOL get_redirect( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *redir );
 extern NTSTATUS get_nt_and_unix_names( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *nt_name,
                                        char **unix_name, UINT disposition, BOOL open_reparse );
-extern NTSTATUS unix_to_nt_file_name( const char *unix_name, WCHAR **nt, UINT disposition );
-extern NTSTATUS get_full_path( char *name, const WCHAR *curdir, UNICODE_STRING *nt_name );
-extern NTSTATUS get_nt_path( const WCHAR *name, UNICODE_STRING *nt_name );
+extern NTSTATUS unix_to_nt_file_name( const char *name, WCHAR **nt );
+extern NTSTATUS get_full_path( const WCHAR *name, const WCHAR *curdir, WCHAR **path );
 extern NTSTATUS open_unix_file( HANDLE *handle, const char *unix_name, ACCESS_MASK access,
                                 OBJECT_ATTRIBUTES *attr, ULONG attributes, ULONG sharing, ULONG disposition,
                                 ULONG options, void *ea_buffer, ULONG ea_length );
 extern NTSTATUS get_device_info( int fd, struct _FILE_FS_DEVICE_INFORMATION *info );
 extern void init_files(void);
+extern void fill_cpu_override(void);
 extern void init_cpu_info(void);
-extern void init_shared_data_cpuinfo( struct _KUSER_SHARED_DATA *data );
 extern void file_complete_async( HANDLE handle, unsigned int options, HANDLE event, PIO_APC_ROUTINE apc, void *apc_user,
                                  IO_STATUS_BLOCK *io, NTSTATUS status, ULONG_PTR information );
 extern void set_async_direct_result( HANDLE *async_handle, unsigned int options, IO_STATUS_BLOCK *io,
                                      NTSTATUS status, ULONG_PTR information, BOOL mark_pending );
+extern struct cpu_topology_override *get_cpu_topology_override(void);
 
 extern NTSTATUS unixcall_wine_dbg_write( void *args );
 extern NTSTATUS unixcall_wine_server_call( void *args );
@@ -388,17 +404,16 @@
 
 extern void dbg_init(void);
 
-extern void close_inproc_sync( HANDLE handle );
-
-extern NTSTATUS call_user_apc_dispatcher( CONTEXT *context_ptr, unsigned int flags, ULONG_PTR arg1, ULONG_PTR arg2,
-                                          ULONG_PTR arg3, PNTAPCFUNC func, NTSTATUS status );
+extern NTSTATUS call_user_apc_dispatcher( CONTEXT *context_ptr, ULONG_PTR arg1, ULONG_PTR arg2, ULONG_PTR arg3,
+                                          PNTAPCFUNC func, NTSTATUS status );
 extern NTSTATUS call_user_exception_dispatcher( EXCEPTION_RECORD *rec, CONTEXT *context );
 extern void call_raise_user_exception_dispatcher(void);
 
 #define IMAGE_DLLCHARACTERISTICS_PREFER_NATIVE 0x0010 /* Wine extension */
 
+extern const char * wine_debuginfostr_pc(void *pc);
+
 #define TICKSPERSEC 10000000
-#define NSECPERSEC 1000000000
 #define SECS_1601_TO_1970  ((369 * 365 + 89) * (ULONGLONG)86400)
 
 static inline ULONGLONG ticks_from_time_t( time_t time )
@@ -432,10 +447,11 @@
             (char *)ptr < (char *)get_signal_stack() + signal_stack_size);
 }
 
-static inline BOOL is_inside_syscall( ULONG_PTR sp )
-{
-    return ((char *)sp >= (char *)ntdll_get_thread_data()->kernel_stack &&
-            (char *)sp <= (char *)get_syscall_frame());
+static inline BOOL is_inside_syscall_stack_guard( const char *stack_ptr )
+{
+    const char *kernel_stack = ntdll_get_thread_data()->kernel_stack;
+
+    return (stack_ptr >= kernel_stack && stack_ptr < kernel_stack + kernel_stack_guard_size);
 }
 
 static inline BOOL is_ec_code( ULONG_PTR ptr )
@@ -470,7 +486,7 @@
 
 static inline NTSTATUS wait_async( HANDLE handle, BOOL alertable )
 {
-    return server_wait_for_object( handle, alertable, NULL );
+    return wait_internal_server( handle, alertable, NULL );
 }
 
 static inline BOOL in_wow64_call(void)
@@ -593,101 +609,4 @@
                                0, NULL, size, ViewShare, 0, protect );
 }
 
-/* LDT definitions */
-
-#if defined(__i386__) || defined(__x86_64__)
-
-struct ldt_bits
-{
-    unsigned int limit : 24;
-    unsigned int type : 5;
-    unsigned int granularity : 1;
-    unsigned int default_big : 1;
-};
-
-#define LDT_SIZE 8192
-
-extern UINT ldt_bitmap[LDT_SIZE / 32];
-
-extern void ldt_set_entry( WORD sel, LDT_ENTRY entry );
-extern WORD ldt_update_entry( WORD sel, LDT_ENTRY entry );
-extern NTSTATUS ldt_get_entry( WORD sel, CLIENT_ID client_id, LDT_ENTRY *entry );
-
-static const LDT_ENTRY null_entry;
-static const struct ldt_bits data_segment = { .type = 0x13, .default_big = 1 };
-static const struct ldt_bits code_segment = { .type = 0x1b, .default_big = 1 };
-
-static inline unsigned int ldt_get_base( LDT_ENTRY ent )
-{
-    return (ent.BaseLow |
-            (unsigned int)ent.HighWord.Bits.BaseMid << 16 |
-            (unsigned int)ent.HighWord.Bits.BaseHi << 24);
-}
-
-static inline struct ldt_bits ldt_set_limit( struct ldt_bits bits, unsigned int limit )
-{
-    if ((bits.granularity = (limit >= 0x100000))) limit >>= 12;
-    bits.limit = limit;
-    return bits;
-}
-
-static inline LDT_ENTRY ldt_make_entry( unsigned int base, struct ldt_bits bits )
-{
-    LDT_ENTRY entry;
-
-    entry.BaseLow                   = (WORD)base;
-    entry.HighWord.Bits.BaseMid     = (BYTE)(base >> 16);
-    entry.HighWord.Bits.BaseHi      = (BYTE)(base >> 24);
-    entry.LimitLow                  = (WORD)bits.limit;
-    entry.HighWord.Bits.LimitHi     = bits.limit >> 16;
-    entry.HighWord.Bits.Dpl         = 3;
-    entry.HighWord.Bits.Pres        = 1;
-    entry.HighWord.Bits.Type        = bits.type;
-    entry.HighWord.Bits.Granularity = bits.granularity;
-    entry.HighWord.Bits.Sys         = 0;
-    entry.HighWord.Bits.Reserved_0  = 0;
-    entry.HighWord.Bits.Default_Big = bits.default_big;
-    return entry;
-}
-
-static inline WORD ldt_alloc_entry( LDT_ENTRY entry )
-{
-    int idx;
-    for (idx = 0; idx < ARRAY_SIZE(ldt_bitmap); idx++)
-    {
-        if (ldt_bitmap[idx] == ~0u) continue;
-        idx = idx * 32 + ffs( ~ldt_bitmap[idx] ) - 1;
-        return ldt_update_entry( (idx << 3) | 7, entry );
-    }
-    return 0;
-}
-
-static inline void ldt_free_entry( WORD sel )
-{
-    unsigned int index = sel >> 3;
-    ldt_bitmap[index / 32] &= ~(1u << (index & 31));
-}
-
-static inline LDT_ENTRY ldt_make_cs32_entry(void)
-{
-    return ldt_make_entry( 0, ldt_set_limit( code_segment, ~0u ));
-}
-
-static inline LDT_ENTRY ldt_make_ds32_entry(void)
-{
-    return ldt_make_entry( 0, ldt_set_limit( data_segment, ~0u ));
-}
-
-static inline LDT_ENTRY ldt_make_fs32_entry( void *teb )
-{
-    return ldt_make_entry( PtrToUlong(teb), ldt_set_limit( data_segment, page_size - 1 ));
-}
-
-static inline int is_gdt_sel( WORD sel )
-{
-    return !(sel & 4);
-}
-
-#endif  /* defined(__i386__) || defined(__x86_64__) */
-
 #endif /* __NTDLL_UNIX_PRIVATE_H */
