diff --git a/dlls/mfreadwrite/reader.c b/dlls/mfreadwrite/reader.c
--- a/dlls/mfreadwrite/reader.c
+++ b/dlls/mfreadwrite/reader.c
@@ -89,7 +89,6 @@
     IMFMediaStream *stream;
     IMFMediaType *current;
     struct list transforms;
-    IMFVideoSampleAllocatorEx *allocator;
     IMFTransform *transform_service;
     DWORD id;
     unsigned int index;
@@ -176,6 +175,9 @@
     CONDITION_VARIABLE sample_event;
     CONDITION_VARIABLE state_event;
     CONDITION_VARIABLE stop_event;
+
+    BOOL flag_eos_for_all_streams;
+    DWORD next_stream_eos_index;
 };
 
 static inline struct source_reader *impl_from_IMFSourceReaderEx(IMFSourceReaderEx *iface)
@@ -232,8 +234,6 @@
         IMFMediaStream_Release(stream->stream);
     if (stream->current)
         IMFMediaType_Release(stream->current);
-    if (stream->allocator)
-        IMFVideoSampleAllocatorEx_Release(stream->allocator);
 }
 
 static ULONG source_reader_release(struct source_reader *reader)
@@ -427,39 +427,6 @@
         WakeAllConditionVariable(&reader->sample_event);
 
     stream->requests--;
-}
-
-static void source_reader_copy_sample_buffer(IMFSample *src, IMFSample *dst)
-{
-    IMFMediaBuffer *buffer;
-    LONGLONG time;
-    DWORD flags;
-    HRESULT hr;
-
-    IMFSample_CopyAllItems(src, (IMFAttributes *)dst);
-
-    IMFSample_SetSampleDuration(dst, 0);
-    IMFSample_SetSampleTime(dst, 0);
-    IMFSample_SetSampleFlags(dst, 0);
-
-    if (SUCCEEDED(IMFSample_GetSampleDuration(src, &time)))
-        IMFSample_SetSampleDuration(dst, time);
-
-    if (SUCCEEDED(IMFSample_GetSampleTime(src, &time)))
-        IMFSample_SetSampleTime(dst, time);
-
-    if (SUCCEEDED(IMFSample_GetSampleFlags(src, &flags)))
-        IMFSample_SetSampleFlags(dst, flags);
-
-    if (SUCCEEDED(IMFSample_ConvertToContiguousBuffer(src, NULL)))
-    {
-        if (SUCCEEDED(IMFSample_GetBufferByIndex(dst, 0, &buffer)))
-        {
-            if (FAILED(hr = IMFSample_CopyToBuffer(src, buffer)))
-                WARN("Failed to copy a buffer, hr %#lx.\n", hr);
-            IMFMediaBuffer_Release(buffer);
-        }
-    }
 }
 
 static HRESULT source_reader_queue_response(struct source_reader *reader, struct media_stream *stream, HRESULT status,
@@ -680,26 +647,31 @@
     return source_reader_release(reader);
 }
 
-static HRESULT source_reader_allocate_stream_sample(MFT_OUTPUT_STREAM_INFO *info, IMFSample **out)
-{
+static HRESULT source_reader_allocate_stream_sample(IMFTransform *transform, MFT_OUTPUT_STREAM_INFO *info, IMFSample **out)
+{
+    IMFMediaType *media_type;
     IMFMediaBuffer *buffer;
     IMFSample *sample;
     HRESULT hr;
 
     *out = NULL;
-    if (FAILED(hr = MFCreateSample(&sample)))
+    if (SUCCEEDED(hr = IMFTransform_GetOutputCurrentType(transform, 0, &media_type)))
+    {
+        hr = MFCreateMediaBufferFromMediaType(media_type, 10000000, info->cbSize, info->cbAlignment, &buffer);
+        IMFMediaType_Release(media_type);
+    }
+    if (FAILED(hr) && FAILED(hr = MFCreateAlignedMemoryBuffer(info->cbSize, info->cbAlignment, &buffer)))
         return hr;
-    if (SUCCEEDED(hr = MFCreateAlignedMemoryBuffer(info->cbSize, info->cbAlignment, &buffer)))
+
+    if (SUCCEEDED(hr = MFCreateSample(&sample)))
     {
         if (SUCCEEDED(hr = IMFSample_AddBuffer(sample, buffer)))
-        {
             *out = sample;
-            IMFSample_AddRef(sample);
-        }
-        IMFMediaBuffer_Release(buffer);
-    }
-
-    IMFSample_Release(sample);
+        else
+            IMFSample_Release(sample);
+    }
+
+    IMFMediaBuffer_Release(buffer);
     return hr;
 }
 
@@ -887,7 +859,7 @@
         IMFMediaType *media_type;
 
         if (!(stream_info.dwFlags & (MFT_OUTPUT_STREAM_PROVIDES_SAMPLES | MFT_OUTPUT_STREAM_CAN_PROVIDE_SAMPLES))
-                && FAILED(hr = source_reader_allocate_stream_sample(&stream_info, &out_buffer.pSample)))
+                && FAILED(hr = source_reader_allocate_stream_sample(entry->transform, &stream_info, &out_buffer.pSample)))
             break;
 
         if (SUCCEEDED(hr = IMFTransform_ProcessOutput(entry->transform, 0, 1, &out_buffer, &status)))
@@ -1221,8 +1193,6 @@
 static struct stream_response *media_stream_pop_response(struct source_reader *reader, struct media_stream *stream)
 {
     struct stream_response *response;
-    IMFSample *sample;
-    HRESULT hr;
 
     LIST_FOR_EACH_ENTRY(response, &reader->responses, struct stream_response, entry)
     {
@@ -1230,26 +1200,6 @@
             continue;
 
         if (!stream) stream = &reader->streams[response->stream_index];
-
-        if (response->sample && stream->allocator)
-        {
-            /* Return allocation error to the caller, while keeping original response sample in for later. */
-            if (SUCCEEDED(hr = IMFVideoSampleAllocatorEx_AllocateSample(stream->allocator, &sample)))
-            {
-                source_reader_copy_sample_buffer(response->sample, sample);
-                IMFSample_Release(response->sample);
-                response->sample = sample;
-            }
-            else
-            {
-                if (!(response = calloc(1, sizeof(*response))))
-                    return NULL;
-
-                response->status = hr;
-                response->stream_flags = MF_SOURCE_READERF_ERROR;
-                return response;
-            }
-        }
 
         return media_stream_detach_response(reader, response);
     }
@@ -1399,9 +1349,11 @@
     }
 
     /* If all selected streams reached EOS, use first selected. */
-    if (first_selected != ~0u)
-    {
-        if (min_ts == MAXLONGLONG)
+    if (first_selected != ~0u && min_ts == MAXLONGLONG)
+    {
+        if (reader->flag_eos_for_all_streams)
+            *stream_index = reader->next_stream_eos_index++ % reader->stream_count;
+        else
             *stream_index = first_selected;
     }
 
@@ -1664,7 +1616,6 @@
 {
     struct source_reader *reader = impl_from_IMFSourceReaderEx(iface);
     ULONG refcount = InterlockedDecrement(&reader->public_refcount);
-    unsigned int i;
 
     TRACE("%p, refcount %lu.\n", iface, refcount);
 
@@ -1684,23 +1635,6 @@
             LeaveCriticalSection(&reader->cs);
         }
 
-        for (i = 0; i < reader->stream_count; ++i)
-        {
-            struct media_stream *stream = &reader->streams[i];
-            IMFVideoSampleAllocatorCallback *callback;
-
-            if (!stream->allocator)
-                continue;
-
-            if (SUCCEEDED(IMFVideoSampleAllocatorEx_QueryInterface(stream->allocator, &IID_IMFVideoSampleAllocatorCallback,
-                    (void **)&callback)))
-            {
-                IMFVideoSampleAllocatorCallback_SetCallback(callback, NULL);
-                IMFVideoSampleAllocatorCallback_Release(callback);
-            }
-        }
-
-        MFUnlockWorkQueue(reader->queue);
         source_reader_release(reader);
     }
 
@@ -1951,71 +1885,6 @@
     return type_set ? S_OK : S_FALSE;
 }
 
-static HRESULT source_reader_create_sample_allocator_attributes(const struct source_reader *reader,
-        IMFAttributes **attributes)
-{
-    UINT32 shared = 0, shared_without_mutex = 0;
-    HRESULT hr;
-
-    if (FAILED(hr = MFCreateAttributes(attributes, 1)))
-        return hr;
-
-    IMFAttributes_GetUINT32(reader->attributes, &MF_SA_D3D11_SHARED, &shared);
-    IMFAttributes_GetUINT32(reader->attributes, &MF_SA_D3D11_SHARED_WITHOUT_MUTEX, &shared_without_mutex);
-
-    if (shared_without_mutex)
-        hr = IMFAttributes_SetUINT32(*attributes, &MF_SA_D3D11_SHARED_WITHOUT_MUTEX, TRUE);
-    else if (shared)
-        hr = IMFAttributes_SetUINT32(*attributes, &MF_SA_D3D11_SHARED, TRUE);
-
-    return hr;
-}
-
-static HRESULT source_reader_setup_sample_allocator(struct source_reader *reader, unsigned int index)
-{
-    struct media_stream *stream = &reader->streams[index];
-    IMFAttributes *attributes = NULL;
-    GUID major = { 0 };
-    HRESULT hr;
-
-    IMFMediaType_GetMajorType(stream->current, &major);
-    if (!IsEqualGUID(&major, &MFMediaType_Video))
-        return S_OK;
-
-    if (!(reader->flags & SOURCE_READER_HAS_DEVICE_MANAGER))
-        return S_OK;
-
-    if (!stream->allocator)
-    {
-        if (FAILED(hr = MFCreateVideoSampleAllocatorEx(&IID_IMFVideoSampleAllocatorEx, (void **)&stream->allocator)))
-        {
-            WARN("Failed to create sample allocator, hr %#lx.\n", hr);
-            return hr;
-        }
-    }
-
-    IMFVideoSampleAllocatorEx_UninitializeSampleAllocator(stream->allocator);
-    if (FAILED(hr = IMFVideoSampleAllocatorEx_SetDirectXManager(stream->allocator, reader->device_manager)))
-    {
-        WARN("Failed to set device manager, hr %#lx.\n", hr);
-        return hr;
-    }
-
-    if (FAILED(hr = source_reader_create_sample_allocator_attributes(reader, &attributes)))
-        WARN("Failed to create allocator attributes, hr %#lx.\n", hr);
-
-    if (FAILED(hr = IMFVideoSampleAllocatorEx_InitializeSampleAllocatorEx(stream->allocator, 2, 8,
-            attributes, stream->current)))
-    {
-        WARN("Failed to initialize sample allocator, hr %#lx.\n", hr);
-    }
-
-    if (attributes)
-        IMFAttributes_Release(attributes);
-
-    return hr;
-}
-
 static BOOL source_reader_allow_video_processor(struct source_reader *reader, BOOL *advanced)
 {
     UINT32 value;
@@ -2108,12 +1977,21 @@
     list_init(&entry->entry);
     entry->category = category;
 
-    if (IsEqualGUID(&out_type.guidMajorType, &MFMediaType_Audio)
-            && SUCCEEDED(IMFMediaType_GetUINT32(output_type, &MF_MT_AUDIO_BLOCK_ALIGNMENT, &entry->min_buffer_size)))
+    if (IsEqualGUID(&out_type.guidMajorType, &MFMediaType_Audio))
     {
         UINT32 bytes_per_second;
 
-        if (SUCCEEDED(IMFMediaType_GetUINT32(output_type, &MF_MT_AUDIO_AVG_BYTES_PER_SECOND, &bytes_per_second)))
+        /* decoders require to have MF_MT_AUDIO_BITS_PER_SAMPLE attribute set, but the source reader doesn't */
+        if (FAILED(IMFMediaType_GetItem(output_type, &MF_MT_AUDIO_BITS_PER_SAMPLE, NULL)))
+        {
+            if (IsEqualGUID(&out_type.guidSubtype, &MFAudioFormat_PCM))
+                IMFMediaType_SetUINT32(output_type, &MF_MT_AUDIO_BITS_PER_SAMPLE, 16);
+            else if (IsEqualGUID(&out_type.guidSubtype, &MFAudioFormat_Float))
+                IMFMediaType_SetUINT32(output_type, &MF_MT_AUDIO_BITS_PER_SAMPLE, 32);
+        }
+
+        if (SUCCEEDED(IMFMediaType_GetUINT32(output_type, &MF_MT_AUDIO_BLOCK_ALIGNMENT, &entry->min_buffer_size))
+                && SUCCEEDED(IMFMediaType_GetUINT32(output_type, &MF_MT_AUDIO_AVG_BYTES_PER_SECOND, &bytes_per_second)))
             entry->min_buffer_size = max(entry->min_buffer_size, bytes_per_second);
     }
 
@@ -2127,6 +2005,15 @@
         out_type.guidSubtype = MFVideoFormat_RGB32;
     }
 
+    if (IsEqualGUID(&out_type.guidMajorType, &MFMediaType_Video) && IsEqualGUID(&out_type.guidSubtype, &MFVideoFormat_IYUV)
+            && IsEqualGUID(&category, &MFT_CATEGORY_VIDEO_DECODER))
+    {
+        /* The WMV video decoder isn't registered for MFVideoFormat_IYUV, but selecting it as an output format still succeeds,
+         * the host decoders usually support IYUV as well, so fixup the subtype for MFTEnumEx.
+         */
+        WARN("Fixing up MFVideoFormat_IYUV subtype for the video processor\n");
+        out_type.guidSubtype = MFVideoFormat_NV12;
+    }
 
     count = 0;
     if (SUCCEEDED(hr = MFTEnumEx(category, 0, &in_type, allow_processor ? NULL : &out_type, &activates, &count)))
@@ -2293,6 +2180,7 @@
         IMFMediaType *type)
 {
     struct source_reader *reader = impl_from_IMFSourceReaderEx(iface);
+    IMFMediaType *output_type;
     HRESULT hr;
 
     TRACE("%p, %#lx, %p, %p.\n", iface, index, reserved, type);
@@ -2312,18 +2200,25 @@
     if (index >= reader->stream_count)
         return MF_E_INVALIDSTREAMNUMBER;
 
+    if (FAILED(hr = MFCreateMediaType(&output_type)))
+        return hr;
+    if (FAILED(IMFMediaType_CopyAllItems(type, (IMFAttributes *)output_type)))
+    {
+        IMFMediaType_Release(output_type);
+        return hr;
+    }
+
     /* FIXME: setting the output type while streaming should trigger a flush */
 
     EnterCriticalSection(&reader->cs);
 
-    hr = source_reader_set_compatible_media_type(reader, index, type);
+    hr = source_reader_set_compatible_media_type(reader, index, output_type);
     if (hr == S_FALSE)
-        hr = source_reader_create_decoder_for_stream(reader, index, type);
-    if (SUCCEEDED(hr))
-        hr = source_reader_setup_sample_allocator(reader, index);
+        hr = source_reader_create_decoder_for_stream(reader, index, output_type);
 
     LeaveCriticalSection(&reader->cs);
 
+    IMFMediaType_Release(output_type);
     return hr;
 }
 
@@ -2546,7 +2441,18 @@
     struct source_reader *reader = impl_from_IMFSourceReaderEx(iface);
     HRESULT hr;
 
+    const char *sgi;
+
     TRACE("%p, %#lx.\n", iface, index);
+
+    sgi = getenv("SteamGameId");
+    if (sgi && strcmp(sgi, "1293160") == 0)
+    {
+        /* In The Medium flushes sometimes lead to the callback
+           calling objects that have already been destroyed. */
+        WARN("ignoring flush\n");
+        return S_OK;
+    }
 
     EnterCriticalSection(&reader->cs);
 
@@ -2806,6 +2712,8 @@
     unsigned int i;
     HRESULT hr;
 
+    const char *sgi;
+
     object = calloc(1, sizeof(*object));
     if (!object)
         return E_OUTOFMEMORY;
@@ -2927,6 +2835,16 @@
         }
     }
 
+    if (object->stream_count > 1 && (sgi = getenv("SteamGameId")) && strcmp(sgi, "462780") == 0)
+    {
+        /* Darksiders Warmastered Edition ends media sampling only when MF_SOURCE_READERF_ENDOFSTREAM
+         * is returned for all streams. If audio and video end simultaneously then ENDOFSTREAM is
+         * flagged only for stream 0, therefore the game depends on a slight time difference which
+         * usually does not occur for the fourth splash video. */
+        WARN("HACK: enabled flagging ENDOFSTREAM for all streams.\n");
+        object->flag_eos_for_all_streams = TRUE;
+    }
+
     if (FAILED(hr = MFLockSharedWorkQueue(L"", 0, NULL, &object->queue)))
         WARN("Failed to acquired shared queue, hr %#lx.\n", hr);
 
