diff --git a/dlls/ntdll/unix/server.c b/dlls/ntdll/unix/server.c
--- a/dlls/ntdll/unix/server.c
+++ b/dlls/ntdll/unix/server.c
@@ -79,10 +79,13 @@
 #include "wine/server.h"
 #include "wine/debug.h"
 #include "unix_private.h"
+#include "esync.h"
+#include "fsync.h"
 #include "ddk/wdm.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(server);
-WINE_DECLARE_DEBUG_CHANNEL(syscall);
+WINE_DECLARE_DEBUG_CHANNEL(client);
+WINE_DECLARE_DEBUG_CHANNEL(ftrace);
 
 #ifndef MSG_CMSG_CLOEXEC
 #define MSG_CMSG_CLOEXEC 0
@@ -97,6 +100,7 @@
 USHORT supported_machines[8] = { 0 };
 USHORT native_machine = 0;
 BOOL process_exiting = FALSE;
+ULONG process_cookie;
 
 timeout_t server_start_time = 0;  /* time of server startup */
 
@@ -157,7 +161,7 @@
     va_list args;
 
     va_start( args, err );
-    fprintf( stderr, "wine client error:%x: ", GetCurrentThreadId() );
+    fprintf( stderr, "wine client error:%x: ", (int)GetCurrentThreadId() );
     vfprintf( stderr, err, args );
     va_end( args );
     abort_thread(1);
@@ -169,7 +173,7 @@
  */
 static DECLSPEC_NORETURN void server_protocol_perror( const char *err )
 {
-    fprintf( stderr, "wine client error:%x: ", GetCurrentThreadId() );
+    fprintf( stderr, "wine client error:%x: ", (int)GetCurrentThreadId() );
     perror( err );
     abort_thread(1);
 }
@@ -182,27 +186,18 @@
  */
 static unsigned int send_request( const struct __server_request_info *req )
 {
-    int request_fd = ntdll_get_thread_data()->request_fd;
+    unsigned int i;
+    int ret;
 
     if (!req->u.req.request_header.request_size)
     {
-        data_size_t to_write = sizeof(req->u.req);
-        const char *write_ptr = (const char *)&req->u.req;
-
-        for (;;)
-        {
-            ssize_t ret = write( request_fd, write_ptr, to_write );
-            if (ret == to_write) return STATUS_SUCCESS;
-            if (ret < 0) break;
-            to_write -= ret;
-            write_ptr += ret;
-        }
+        if ((ret = write( ntdll_get_thread_data()->request_fd, &req->u.req,
+                          sizeof(req->u.req) )) == sizeof(req->u.req)) return STATUS_SUCCESS;
+
     }
     else
     {
-        data_size_t to_write = sizeof(req->u.req) + req->u.req.request_header.request_size;
         struct iovec vec[__SERVER_MAX_DATA+1];
-        unsigned int i, j;
 
         vec[0].iov_base = (void *)&req->u.req;
         vec[0].iov_len = sizeof(req->u.req);
@@ -211,30 +206,11 @@
             vec[i+1].iov_base = (void *)req->data[i].ptr;
             vec[i+1].iov_len = req->data[i].size;
         }
-
-        for (;;)
-        {
-            ssize_t ret = writev( request_fd, vec, i + 1 );
-            if (ret == to_write) return STATUS_SUCCESS;
-            if (ret < 0) break;
-            to_write -= ret;
-            for (j = 0; j < i + 1; j++)
-            {
-                if (ret >= vec[j].iov_len)
-                {
-                    ret -= vec[j].iov_len;
-                    vec[j].iov_len = 0;
-                }
-                else
-                {
-                    vec[j].iov_base = (char *)vec[j].iov_base + ret;
-                    vec[j].iov_len -= ret;
-                    break;
-                }
-            }
-        }
-    }
-
+        if ((ret = writev( ntdll_get_thread_data()->request_fd, vec, i+1 )) ==
+            req->u.req.request_header.request_size + sizeof(req->u.req)) return STATUS_SUCCESS;
+    }
+
+    if (ret >= 0) server_protocol_error( "partial write %d\n", ret );
     if (errno == EPIPE) abort_thread(0);
     if (errno == EFAULT) return STATUS_ACCESS_VIOLATION;
     server_protocol_perror( "write" );
@@ -290,8 +266,13 @@
     struct __server_request_info * const req = req_ptr;
     unsigned int ret;
 
-    if ((ret = send_request( req ))) return ret;
-    return wait_reply( req );
+    FTRACE_BLOCK_START("req %s", req->name)
+    TRACE_(client)("%s start\n", req->name); \
+    if (!(ret = send_request( req )))
+        ret = wait_reply( req );
+    TRACE_(client)("%s end\n", req->name);
+    FTRACE_BLOCK_END()
+    return ret;
 }
 
 
@@ -385,7 +366,7 @@
  */
 static NTSTATUS invoke_user_apc( CONTEXT *context, const struct user_apc *apc, NTSTATUS status )
 {
-    return call_user_apc_dispatcher( context, apc->flags, apc->args[0], apc->args[1], apc->args[2],
+    return call_user_apc_dispatcher( context, apc->args[0], apc->args[1], apc->args[2],
                                      wine_server_get_ptr( apc->func ), status );
 }
 
@@ -774,7 +755,9 @@
         pthread_sigmask( SIG_SETMASK, &old_set, NULL );
         if (signaled) break;
 
+        FTRACE_BLOCK_START("select_reply")
         ret = wait_select_reply( &cookie );
+        FTRACE_BLOCK_END()
     }
     while (ret == STATUS_USER_APC || ret == STATUS_KERNEL_APC);
 
@@ -815,21 +798,6 @@
        boost as well.  This seems to model that behavior the closest.  */
     if (ret == STATUS_TIMEOUT) NtYieldExecution();
     return ret;
-}
-
-
-/* helper function to perform a server-side wait on an internal handle without
- * using the fast synchronization path */
-unsigned int server_wait_for_object( HANDLE handle, BOOL alertable, const LARGE_INTEGER *timeout )
-{
-    union select_op select_op;
-    UINT flags = SELECT_INTERRUPTIBLE;
-
-    if (alertable) flags |= SELECT_ALERTABLE;
-
-    select_op.wait.op = SELECT_WAIT;
-    select_op.wait.handles[0] = wine_server_obj_handle( handle );
-    return server_wait( &select_op, offsetof( union select_op, wait.handles[1] ), flags, timeout );
 }
 
 
@@ -909,24 +877,15 @@
         }
         else
         {
-            sigset_t sigset;
-
             NtWaitForSingleObject( handle, FALSE, NULL );
-
-            server_enter_uninterrupted_section( &fd_cache_mutex, &sigset );
-
-            /* remove the handle from the cache, get_apc_result will close it for us */
-            close_inproc_sync( handle );
 
             SERVER_START_REQ( get_apc_result )
             {
                 req->handle = wine_server_obj_handle( handle );
-                if (!(ret = server_call_unlocked( req ))) *result = reply->result;
+                if (!(ret = wine_server_call( req ))) *result = reply->result;
             }
             SERVER_END_REQ;
 
-            server_leave_uninterrupted_section( &fd_cache_mutex, &sigset );
-
             if (!ret && result->type == APC_NONE) continue;  /* APC didn't run, try again */
         }
         return ret;
@@ -939,7 +898,7 @@
  *
  * Send a file descriptor to the server.
  */
-void CDECL wine_server_send_fd( int fd )
+void wine_server_send_fd( int fd )
 {
     struct send_fd data;
     struct msghdr msghdr;
@@ -985,7 +944,7 @@
  *
  * Receive a file descriptor passed from the server.
  */
-int wine_server_receive_fd( obj_handle_t *handle )
+int receive_fd( obj_handle_t *handle )
 {
     struct iovec vec;
     struct msghdr msghdr;
@@ -1180,7 +1139,7 @@
                 if (type) *type = reply->type;
                 if (options) *options = reply->options;
                 access = reply->access;
-                if ((fd = wine_server_receive_fd( &fd_handle )) != -1)
+                if ((fd = receive_fd( &fd_handle )) != -1)
                 {
                     assert( wine_server_ptr_handle(fd_handle) == handle );
                     *needs_close = (!reply->cacheable ||
@@ -1324,11 +1283,6 @@
     char *p;
     struct stat st;
     int fd_cwd = open( ".", O_RDONLY );
-
-#ifdef _SC_PAGESIZE
-    if (sysconf( _SC_PAGESIZE ) != 4096)
-        MESSAGE("You are running with the wrong page size, expect problems!\nPlease also consult https://github.com/AndreRH/hangover/wiki/Page-Size\n");
-#endif
 
     if (chdir( config_dir ) == -1)
     {
@@ -1597,9 +1551,9 @@
  */
 size_t server_init_process(void)
 {
+    struct cpu_topology_override *cpu_override;
     const char *arch = getenv( "WINEARCH" );
     const char *env_socket = getenv( "WINESERVERSOCKET" );
-    struct ntdll_thread_data *data = ntdll_get_thread_data();
     obj_handle_t version;
     unsigned int i;
     int ret, reply_pipe;
@@ -1621,8 +1575,8 @@
 
         if (is_win64 && arch && !strcmp( arch, "win32" ))
             fatal_error( "WINEARCH is set to 'win32' but this is not supported in wow64 mode.\n" );
-        if (arch && strcmp( arch, "win32" ) && strcmp( arch, "win64" ) && strcmp( arch, "wow64" ))
-            fatal_error( "WINEARCH set to invalid value '%s', it must be win32, win64, or wow64.\n", arch );
+        if (arch && strcmp( arch, "win32" ) && strcmp( arch, "win64" ))
+            fatal_error( "WINEARCH set to invalid value '%s', it must be either win32 or win64.\n", arch );
 
         fd_socket = server_connect();
     }
@@ -1639,7 +1593,7 @@
     pthread_sigmask( SIG_BLOCK, &server_block_set, NULL );
 
     /* receive the first thread request fd on the main socket */
-    data->request_fd = wine_server_receive_fd( &version );
+    ntdll_get_thread_data()->request_fd = receive_fd( &version );
 
 #ifdef SO_PASSCRED
     /* now that we hopefully received the server_pid, disable SO_PASSCRED */
@@ -1658,7 +1612,7 @@
                                (version > SERVER_PROTOCOL_VERSION) ? "wine" : "wineserver" );
 #if defined(__linux__) && defined(HAVE_PRCTL)
     /* work around Ubuntu's ptrace breakage */
-    if (server_pid != -1) prctl( 0x59616d61 /* PR_SET_PTRACER */, server_pid );
+    if (server_pid != -1) prctl( 0x59616d61 /* PR_SET_PTRACER */, PR_SET_PTRACER_ANY );
 #endif
 
     /* ignore SIGPIPE so that we get an EPIPE error instead  */
@@ -1669,29 +1623,26 @@
 
     reply_pipe = init_thread_pipe();
 
+    fill_cpu_override();
+    cpu_override = get_cpu_topology_override();
+
     SERVER_START_REQ( init_first_thread )
     {
+        if (cpu_override)
+            wine_server_add_data( req, cpu_override, sizeof(*cpu_override) );
         req->unix_pid    = getpid();
         req->unix_tid    = get_unix_tid();
         req->reply_fd    = reply_pipe;
-        req->wait_fd     = data->wait_fd[1];
+        req->wait_fd     = ntdll_get_thread_data()->wait_fd[1];
         req->debug_level = (TRACE_ON(server) != 0);
         wine_server_set_reply( req, supported_machines, sizeof(supported_machines) );
-        if (!(ret = wine_server_call( req )))
-        {
-            obj_handle_t handle;
-            pid               = reply->pid;
-            tid               = reply->tid;
-            peb->SessionId    = reply->session_id;
-            info_size         = reply->info_size;
-            server_start_time = reply->server_start;
-            supported_machines_count = wine_server_reply_size( reply ) / sizeof(*supported_machines);
-            if (reply->inproc_device)
-            {
-                inproc_device_fd = wine_server_receive_fd( &handle );
-                assert( handle == reply->inproc_device );
-            }
-        }
+        ret = wine_server_call( req );
+        pid               = reply->pid;
+        tid               = reply->tid;
+        peb->SessionId    = reply->session_id;
+        info_size         = reply->info_size;
+        server_start_time = reply->server_start;
+        supported_machines_count = wine_server_reply_size( reply ) / sizeof(*supported_machines);
     }
     SERVER_END_REQ;
     close( reply_pipe );
@@ -1717,11 +1668,13 @@
     {
         if (is_win64)
             fatal_error( "'%s' is a 32-bit installation, it cannot support 64-bit applications.\n", config_dir );
-        if (arch && (!strcmp( arch, "win64" ) || !strcmp( arch, "wow64" )))
-            fatal_error( "WINEARCH set to %s but '%s' is a 32-bit installation.\n", arch, config_dir );
+        if (arch && !strcmp( arch, "win64" ))
+            fatal_error( "WINEARCH set to win64 but '%s' is a 32-bit installation.\n", config_dir );
     }
 
     set_thread_id( NtCurrentTeb(), pid, tid );
+    process_cookie = 0xdeadbeef;
+    get_random( &process_cookie, sizeof(process_cookie) );
 
     for (i = 0; i < supported_machines_count; i++)
         if (supported_machines[i] == current_machine) return info_size;
@@ -1739,7 +1692,6 @@
     unsigned int status;
     int suspend;
     FILE_FS_DEVICE_INFORMATION info;
-    struct ntdll_thread_data *thread_data = ntdll_get_thread_data();
 
     if (!get_device_info( initial_cwd, &info ) && (info.Characteristics & FILE_REMOVABLE_MEDIA))
         chdir( "/" );
@@ -1753,8 +1705,6 @@
      * send exceptions to the debugger before the create process event that
      * is sent by init_process_done */
     signal_init_process();
-    thread_data->syscall_table = KeServiceDescriptorTable;
-    thread_data->syscall_trace = TRACE_ON(syscall);
 
     /* always send the native TEB */
     if (!(teb = NtCurrentTeb64())) teb = NtCurrentTeb();
@@ -1762,8 +1712,11 @@
     /* Signal the parent process to continue */
     SERVER_START_REQ( init_process_done )
     {
-        req->teb = wine_server_client_ptr( teb );
-        req->peb = NtCurrentTeb64() ? NtCurrentTeb64()->Peb : wine_server_client_ptr( peb );
+        req->teb      = wine_server_client_ptr( teb );
+        req->peb      = NtCurrentTeb64() ? NtCurrentTeb64()->Peb : wine_server_client_ptr( peb );
+#ifdef __i386__
+        req->ldt_copy = wine_server_client_ptr( &__wine_ldt_copy );
+#endif
         status = wine_server_call( req );
         suspend = reply->suspend;
     }
@@ -1860,17 +1813,12 @@
         return result.dup_handle.status;
     }
 
-    /* hold fd_cache_mutex to prevent the fd from being added again between the
-     * call to remove_fd_from_cache and close_handle */
     server_enter_uninterrupted_section( &fd_cache_mutex, &sigset );
 
     /* always remove the cached fd; if the server request fails we'll just
      * retrieve it again */
     if (options & DUPLICATE_CLOSE_SOURCE)
-    {
         fd = remove_fd_from_cache( source );
-        close_inproc_sync( source );
-    }
 
     SERVER_START_REQ( dup_handle )
     {
@@ -1936,14 +1884,17 @@
     if (HandleToLong( handle ) >= ~5 && HandleToLong( handle ) <= ~0)
         return STATUS_SUCCESS;
 
-    /* hold fd_cache_mutex to prevent the fd from being added again between the
-     * call to remove_fd_from_cache and close_handle */
     server_enter_uninterrupted_section( &fd_cache_mutex, &sigset );
 
     /* always remove the cached fd; if the server request fails we'll just
      * retrieve it again */
     fd = remove_fd_from_cache( handle );
-    close_inproc_sync( handle );
+
+    if (do_fsync())
+        fsync_close( handle );
+
+    if (do_esync())
+        esync_close( handle );
 
     SERVER_START_REQ( close_handle )
     {
