diff --git a/dlls/setupapi/fakedll.c b/dlls/setupapi/fakedll.c
--- a/dlls/setupapi/fakedll.c
+++ b/dlls/setupapi/fakedll.c
@@ -272,7 +272,6 @@
     DWORD lfanew = (sizeof(*dos) + sizeof(fakedll_signature) + 15) & ~15;
     DWORD size, header_size = lfanew + sizeof(*nt);
 
-    TRACE( "creating %s\n", debugstr_w(name) );
     info.handle = file;
     buffer = calloc( 1, header_size + 8 * sizeof(IMAGE_SECTION_HEADER) );
 
@@ -393,7 +392,7 @@
     while (p != NULL)
     {
         *p = 0;
-        if (!CreateDirectoryW(path, NULL) && GetLastError() != ERROR_ALREADY_EXISTS)
+        if (!CreateDirectoryW(path, NULL))
             TRACE("Couldn't create directory %s - error: %ld\n", wine_dbgstr_w(path), GetLastError());
         *p = '\\';
         p = wcschr(p+1, '\\');
@@ -429,7 +428,7 @@
 }
 
 /* try to load a pre-compiled fake dll */
-static void *load_fake_dll( const WCHAR *name, SIZE_T *size )
+static void *load_fake_dll( const WCHAR *name, SIZE_T *size, WCHAR **out_filename )
 {
     const WCHAR *build_dir = _wgetenv( L"WINEBUILDDIR" );
     const WCHAR *path;
@@ -477,9 +476,92 @@
     }
 
 done:
+    if (res == 1)
+    {
+        *out_filename = malloc( (wcslen( ptr ) + 1) * sizeof(WCHAR) );
+        wcscpy( *out_filename, ptr );
+        free( file );
+        return data;
+    }
     free( file );
-    if (res == 1) return data;
     return NULL;
+}
+
+/* Check if the 2 files are already the same */
+static BOOL fake_dll_matches( const WCHAR *source, const WCHAR *dest )
+{
+    WIN32_FILE_ATTRIBUTE_DATA data1, data2;
+    HANDLE h;
+    char *buf;
+    DWORD size, bytesread;
+    BOOL result;
+
+    if (!GetFileAttributesExW( source, GetFileExInfoStandard, &data1 ))
+        return FALSE;
+
+    if (!GetFileAttributesExW( dest, GetFileExInfoStandard, &data2 ))
+        return FALSE;
+
+    if ((data2.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) == FILE_ATTRIBUTE_REPARSE_POINT)
+        /* If it's a symlink, assume it was put there by the proton script and is therefore correct */
+        return TRUE;
+
+    if (data1.nFileSizeHigh != 0 ||
+        data2.nFileSizeHigh != 0 ||
+        data1.nFileSizeLow != data2.nFileSizeLow)
+    {
+        return FALSE;
+    }
+
+    size = data1.nFileSizeLow;
+
+    /* If size and mtime matches, don't bother comparing contents */
+    if ((data1.ftLastWriteTime.dwLowDateTime != 0 || data1.ftLastWriteTime.dwHighDateTime != 0) &&
+        data1.ftLastWriteTime.dwLowDateTime == data2.ftLastWriteTime.dwLowDateTime &&
+        data1.ftLastWriteTime.dwHighDateTime == data2.ftLastWriteTime.dwHighDateTime)
+        return TRUE;
+
+    buf = malloc( size * 2 );
+
+    if (!buf)
+        return FALSE;
+
+    h = CreateFileW( source, GENERIC_READ, 0, NULL, OPEN_EXISTING, 0, NULL );
+    if (h == INVALID_HANDLE_VALUE)
+    {
+        free( buf );
+        return FALSE;
+    }
+
+    if (!ReadFile( h, buf, size, &bytesread, NULL ) || bytesread != size)
+    {
+        CloseHandle( h );
+        free( buf );
+        return FALSE;
+    }
+
+    CloseHandle( h );
+
+    h = CreateFileW( dest, GENERIC_READ, 0, NULL, OPEN_EXISTING, 0, NULL );
+    if (h == INVALID_HANDLE_VALUE)
+    {
+        free( buf );
+        return FALSE;
+    }
+
+    if (!ReadFile( h, buf + size, size, &bytesread, NULL ) || bytesread != size)
+    {
+        CloseHandle( h );
+        free( buf );
+        return FALSE;
+    }
+
+    CloseHandle( h );
+
+    result = !memcmp( buf, buf + size, size );
+
+    free( buf );
+    return result;
 }
 
 /* create the fake dll destination file */
@@ -504,35 +586,6 @@
         /* truncate the file */
         SetFilePointer( h, 0, NULL, FILE_BEGIN );
         SetEndOfFile( h );
-    }
-    else if (GetLastError() == ERROR_SHARING_VIOLATION &&
-             (h = CreateFileW( name, GENERIC_READ, FILE_SHARE_READ, NULL,
-                               OPEN_EXISTING, 0, NULL )) != INVALID_HANDLE_VALUE)
-    {
-        if (!is_fake_dll( h ))
-        {
-            TRACE( "%s is not a fake dll, not overwriting it\n", debugstr_w(name) );
-            CloseHandle( h );
-            return 0;
-        }
-        CloseHandle( h );
-        h = INVALID_HANDLE_VALUE;
-
-        if (!delete)
-        {
-            WCHAR *p, *path = wcsdup( name );
-            WCHAR tmpname[MAX_PATH];
-
-            if ((p = wcsrchr( path, '\\' ))) *p = 0;
-            if (GetTempFileNameW( path, L"dll", 0, tmpname ))
-            {
-                h = CreateFileW( tmpname, GENERIC_READ|GENERIC_WRITE, 0, NULL, TRUNCATE_EXISTING, 0, NULL );
-                if (h != INVALID_HANDLE_VALUE)
-                    MoveFileExW( tmpname, name, MOVEFILE_DELAY_UNTIL_REBOOT | MOVEFILE_REPLACE_EXISTING );
-            }
-            free( path );
-        }
-        else MoveFileExW( name, NULL, MOVEFILE_DELAY_UNTIL_REBOOT );
     }
     else if (!delete)
     {
@@ -667,40 +720,22 @@
     return TRUE;
 }
 
-static unsigned int append_string( WCHAR *buffer, const xmlstr_t *str, unsigned int maxlen )
-{
-    static const char valid_chars[] = "-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
-    unsigned int i, len;
-
-    if (!str->len) return 0;
-    for (i = len = 0; i < str->len; i++)
-        if (strchr( valid_chars, str->ptr[i] )) buffer[len++] = (WCHAR)str->ptr[i];
-    if (len > maxlen)
-    {
-        unsigned int pos = maxlen / 2;
-        buffer[pos - 1] = buffer[pos] = '.';
-        memmove( buffer + pos + 1, buffer + len - pos + 1, (pos - 1) * sizeof(WCHAR) );
-        len = maxlen;
-    }
-    buffer[len++] = '_';
-    buffer[len] = 0;
-    wcslwr( buffer );
-    return len;
-}
-
 static void append_manifest_filename( const xmlstr_t *arch, const xmlstr_t *name, const xmlstr_t *key,
                                       const xmlstr_t *version, const xmlstr_t *lang, WCHAR *buffer, DWORD size )
 {
     DWORD pos = lstrlenW( buffer );
 
-    pos += append_string( buffer + pos, arch, 16 );
-    pos += append_string( buffer + pos, name, 40 );
+    pos += MultiByteToWideChar( CP_UTF8, 0, arch->ptr, arch->len, buffer + pos, size - pos );
+    buffer[pos++] = '_';
+    pos += MultiByteToWideChar( CP_UTF8, 0, name->ptr, name->len, buffer + pos, size - pos );
+    buffer[pos++] = '_';
     pos += MultiByteToWideChar( CP_UTF8, 0, key->ptr, key->len, buffer + pos, size - pos );
     buffer[pos++] = '_';
     pos += MultiByteToWideChar( CP_UTF8, 0, version->ptr, version->len, buffer + pos, size - pos );
     buffer[pos++] = '_';
-    pos += append_string( buffer + pos, lang, 8 );
-    lstrcpyW( buffer + pos, L"deadbeef" );
+    pos += MultiByteToWideChar( CP_UTF8, 0, lang->ptr, lang->len, buffer + pos, size - pos );
+    lstrcpyW( buffer + pos, L"_deadbeef" );
+    wcslwr( buffer );
 }
 
 static WCHAR* create_winsxs_dll_path( const xmlstr_t *arch, const xmlstr_t *name,
@@ -761,20 +796,21 @@
     struct list entry;
     WCHAR *src;
     WCHAR *dest;
-    WCHAR  data[];
+    WCHAR data[1];
 };
 
 struct dll_data
 {
     struct list *delay_copy;
-    const WCHAR *src;
+    const WCHAR *src_dir;
+    DWORD src_len;
 };
 
 static BOOL CALLBACK register_manifest( HMODULE module, const WCHAR *type, WCHAR *res_name, LONG_PTR arg )
 {
     const struct dll_data *dll_data = (const struct dll_data*)arg;
-    WCHAR *p, *dest = NULL;
-    DWORD src_len = wcslen( dll_data->src ), dest_len = 0, file_count = 0;
+    WCHAR *dest = NULL;
+    DWORD dest_len = 0;
     xmlbuf_t buffer;
     xmlstr_t elem, attr_name, attr_value;
     xmlstr_t name, version, arch, key, lang;
@@ -809,25 +845,18 @@
 
             if (!error && dest && name.ptr)
             {
-                int namelen = MultiByteToWideChar( CP_UTF8, 0, name.ptr, name.len, NULL, 0 );
-                struct delay_copy *add = malloc( offsetof( struct delay_copy,
-                                                           data[src_len + 1 + dest_len + 2 * namelen + 1] ));
+                struct delay_copy *add = malloc( sizeof(*add) +
+                        (dll_data->src_len + name.len + dest_len + name.len + 1) * sizeof(WCHAR) );
                 add->src = add->data;
-                memcpy( add->src, dll_data->src, src_len * sizeof(WCHAR) );
-                if (!file_count++)
-                {
-                    add->src[src_len] = 0;
-                    if ((p = wcsrchr( dll_data->src, '\\' ))) src_len = p + 1 - dll_data->src;
-                }
-                else
-                {
-                    MultiByteToWideChar( CP_UTF8, 0, name.ptr, name.len, add->src + src_len, namelen );
-                    add->src[src_len + namelen] = 0;
-                }
-                add->dest = add->src + wcslen(add->src) + 1;
+                memcpy( add->src, dll_data->src_dir, dll_data->src_len * sizeof(WCHAR) );
+                MultiByteToWideChar( CP_UTF8, 0, name.ptr, name.len,
+                        add->src + dll_data->src_len, name.len );
+                add->src[dll_data->src_len + name.len] = 0;
+                add->dest = add->data + dll_data->src_len + name.len + 1;
                 memcpy( add->dest, dest, dest_len * sizeof(WCHAR) );
-                MultiByteToWideChar( CP_UTF8, 0, name.ptr, name.len, add->dest + dest_len, namelen );
-                add->dest[dest_len + namelen] = 0;
+                memcpy( add->dest + dest_len, add->src + dll_data->src_len,
+                        (name.len + 1) * sizeof(WCHAR) );
+                TRACE("schedule copy %s -> %s\n", wine_dbgstr_w(add->src), wine_dbgstr_w(add->dest));
                 list_add_tail( dll_data->delay_copy, &add->entry );
             }
             continue;
@@ -900,9 +929,13 @@
     LDR_RESOURCE_INFO info;
     HRESULT hr = S_OK;
     HMODULE module = (HMODULE)((ULONG_PTR)data | 1);
-    struct dll_data dll_data = { delay_copy, name };
+    struct dll_data dll_data = { delay_copy, name, 0 };
     WCHAR buffer[MAX_PATH];
-
+    const WCHAR *p;
+
+    if (!(p = wcsrchr( name, '\\' ))) p = name;
+    else p++;
+    dll_data.src_len = p - name;
     EnumResourceNamesW( module, (WCHAR*)RT_MANIFEST, register_manifest, (LONG_PTR)&dll_data );
 
     info.Type = (ULONG_PTR)L"WINE_REGISTRY";
@@ -957,7 +990,11 @@
     destname[len] = 0;
     if (!add_handled_dll( destname )) ret = -1;
 
-    if (ret != -1)
+    if (ret != -1 && fake_dll_matches( file, dest ))
+    {
+        register_fake_dll( dest, data, size, delay_copy );
+    }
+    else if (ret != -1)
     {
         HANDLE h = create_dest_file( dest, delete );
 
@@ -989,6 +1026,13 @@
     LIST_FOR_EACH_ENTRY_SAFE( copy, next, delay_copy, struct delay_copy, entry )
     {
         list_remove( &copy->entry );
+
+        if ( fake_dll_matches( copy->src, copy->dest ) )
+        {
+            free( copy );
+            continue;
+        }
+
         ret = read_file( copy->src, &data, &size );
         if (ret != 1)
         {
@@ -1028,6 +1072,7 @@
         if (lstrlenW( data.name ) > max_dll_name_len) continue;
         if (!wcscmp( data.name, L"." )) continue;
         if (!wcscmp( data.name, L".." )) continue;
+        if (!wcscmp( data.name, L"amd_ags_x64.dll" )) continue;
         lstrcpyW( name, data.name );
         if (default_ext)  /* inside build dir */
         {
@@ -1097,9 +1142,10 @@
 {
     struct list delay_copy = LIST_INIT( delay_copy );
     HANDLE h;
-    BOOL ret = FALSE;
+    BOOL ret;
     SIZE_T size;
     const WCHAR *filename;
+    WCHAR *source_filename;
     void *buffer;
     BOOL delete = !wcscmp( source, L"-" );  /* '-' source means delete the file */
 
@@ -1116,24 +1162,55 @@
 
     add_handled_dll( filename );
 
-    if (!(h = create_dest_file( name, delete ))) return TRUE;  /* not a fake dll */
-    if (h == INVALID_HANDLE_VALUE) return FALSE;
-
-    if (!wcscmp( source, L"*" ))  /* '*' source means create fake file for a non-existent dll */
-    {
+    if (delete)
+    {
+        if (!(h = create_dest_file( name, delete ))) return TRUE;  /* not a fake dll */
+        if (h == INVALID_HANDLE_VALUE) return FALSE;
+
+        /* '-' source means delete the file */
+        TRACE( "deleting %s\n", debugstr_w(name) );
+        ret = FALSE;
+
+        CloseHandle( h );
+        DeleteFileW( name );
+    }
+    else if ((buffer = load_fake_dll( source, &size, &source_filename )))
+    {
+        if (fake_dll_matches( source_filename, name ))
+        {
+            free( source_filename );
+
+            register_fake_dll( name, buffer, size, &delay_copy );
+            ret = TRUE;
+        }
+        else
+        {
+            DWORD written;
+
+            free( source_filename );
+
+            if (!(h = create_dest_file( name, FALSE ))) return TRUE;  /* not a fake dll */
+            if (h == INVALID_HANDLE_VALUE) return FALSE;
+
+            ret = (WriteFile( h, buffer, size, &written, NULL ) && written == size);
+            if (ret) register_fake_dll( name, buffer, size, &delay_copy );
+            else ERR( "failed to write to %s (error=%lu)\n", debugstr_w(name), GetLastError() );
+
+            CloseHandle( h );
+            if (!ret) DeleteFileW( name );
+        }
+    }
+    else
+    {
+        if (!(h = create_dest_file( name, FALSE ))) return TRUE;  /* not a fake dll */
+        if (h == INVALID_HANDLE_VALUE) return FALSE;
+
+        WARN( "fake dll %s not found for %s\n", debugstr_w(source), debugstr_w(name) );
         ret = build_fake_dll( h, name );
-    }
-    else if ((buffer = load_fake_dll( source, &size )))
-    {
-        DWORD written;
-
-        ret = (WriteFile( h, buffer, size, &written, NULL ) && written == size);
-        if (ret) register_fake_dll( name, buffer, size, &delay_copy );
-        else ERR( "failed to write to %s (error=%lu)\n", debugstr_w(name), GetLastError() );
-    }
-
-    CloseHandle( h );
-    if (!ret) DeleteFileW( name );
+
+        CloseHandle( h );
+        if (!ret) DeleteFileW( name );
+    }
 
     delay_copy_files( &delay_copy );
     return ret;
