diff --git a/dlls/ntoskrnl.exe/pnp.c b/dlls/ntoskrnl.exe/pnp.c
--- a/dlls/ntoskrnl.exe/pnp.c
+++ b/dlls/ntoskrnl.exe/pnp.c
@@ -27,6 +27,7 @@
 #include "cfgmgr32.h"
 #include "dbt.h"
 #include "wine/exception.h"
+#include "wine/heap.h"
 
 #include "plugplay.h"
 
@@ -35,12 +36,6 @@
 #include "devpkey.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(plugplay);
-
-DECLARE_CRITICAL_SECTION(invalidated_devices_cs);
-static CONDITION_VARIABLE invalidated_devices_cv = CONDITION_VARIABLE_INIT;
-
-static DEVICE_OBJECT **invalidated_devices;
-static size_t invalidated_devices_count;
 
 static inline const char *debugstr_propkey( const DEVPROPKEY *id )
 {
@@ -332,42 +327,6 @@
     return TRUE;
 }
 
-static void create_dyn_data_key( DEVICE_OBJECT *device )
-{
-    struct wine_device *wine_device = CONTAINING_RECORD(device, struct wine_device, device_obj);
-    WCHAR device_instance_id[MAX_DEVICE_ID_LEN];
-    static unsigned int counter;
-    WCHAR key_path[29];
-    DWORD disposition;
-    LSTATUS ret;
-    HKEY key;
-
-    if (get_device_instance_id( device, device_instance_id ))
-        return;
-
-    for (;;)
-    {
-        swprintf( key_path, ARRAY_SIZE(key_path), L"Config Manager\\Enum\\%08x", counter++ );
-
-        if ((ret = RegCreateKeyExW( HKEY_DYN_DATA, key_path, 0, NULL,
-                REG_OPTION_VOLATILE, KEY_ALL_ACCESS, NULL, &key, &disposition )))
-        {
-            ERR( "Failed to create %s, error %lu.\n", debugstr_w(key_path), GetLastError() );
-            return;
-        }
-
-        if (disposition == REG_CREATED_NEW_KEY)
-        {
-            RegSetValueExW( key, L"HardWareKey", 0, REG_SZ, (BYTE *)device_instance_id,
-                    wcslen( device_instance_id ) * sizeof(WCHAR) );
-            wine_device->dyn_data_key = key;
-            break;
-        }
-
-        RegCloseKey( key );
-    }
-}
-
 /* Load the function driver for a newly created PDO, if one is present, and
  * send IRPs to start the device. */
 static void start_device( DEVICE_OBJECT *device, HDEVINFO set, SP_DEVINFO_DATA *sp_device )
@@ -375,8 +334,6 @@
     load_function_driver( device, set, sp_device );
     if (device->DriverObject)
         send_pnp_irp( device, IRP_MN_START_DEVICE );
-
-    create_dyn_data_key( device );
 }
 
 static void enumerate_new_device( DEVICE_OBJECT *device, HDEVINFO set )
@@ -461,15 +418,6 @@
 
 static void remove_device( DEVICE_OBJECT *device )
 {
-    struct wine_device *wine_device = CONTAINING_RECORD(device, struct wine_device, device_obj);
-
-    if (wine_device->dyn_data_key)
-    {
-        RegDeleteKeyW( wine_device->dyn_data_key, L"" );
-        RegCloseKey( wine_device->dyn_data_key );
-        wine_device->dyn_data_key = 0;
-    }
-
     send_remove_device_irp( device, IRP_MN_SURPRISE_REMOVAL );
     send_remove_device_irp( device, IRP_MN_REMOVE_DEVICE );
 }
@@ -570,14 +518,8 @@
     switch (type)
     {
         case BusRelations:
-            EnterCriticalSection( &invalidated_devices_cs );
-            invalidated_devices = realloc( invalidated_devices,
-                    (invalidated_devices_count + 1) * sizeof(*invalidated_devices) );
-            invalidated_devices[invalidated_devices_count++] = device_object;
-            LeaveCriticalSection( &invalidated_devices_cs );
-            WakeConditionVariable( &invalidated_devices_cv );
-            break;
-
+            handle_bus_relations( device_object );
+            break;
         default:
             FIXME("Unhandled relation %#x.\n", type);
             break;
@@ -654,13 +596,14 @@
         {
             WCHAR *id, *ptr;
 
-            status = get_device_id( device, BusQueryDeviceID, &id );
+            status = get_device_id( device, BusQueryInstanceID, &id );
             if (status != STATUS_SUCCESS)
             {
                 ERR("Failed to get instance ID, status %#lx.\n", status);
                 break;
             }
 
+            wcsupr( id );
             ptr = wcschr( id, '\\' );
             if (ptr) *ptr = 0;
 
@@ -795,28 +738,28 @@
     if (ret != STATUS_BUFFER_TOO_SMALL)
         return ret;
 
-    device_name = malloc( len );
+    device_name = heap_alloc( len );
     ret = IoGetDeviceProperty( device, DevicePropertyPhysicalDeviceObjectName, len, device_name, &len );
     if (ret)
     {
-        free( device_name );
+        heap_free( device_name );
         return ret;
     }
 
     RtlInitUnicodeString( &device_nameU, device_name );
     ret = IoCreateSymbolicLink( symlink_name, &device_nameU );
-    free( device_name );
+    heap_free( device_name );
     return ret;
 }
 
 void  __RPC_FAR * __RPC_USER MIDL_user_allocate( SIZE_T len )
 {
-    return HeapAlloc( GetProcessHeap(), 0, len );
+    return heap_alloc( len );
 }
 
 void __RPC_USER MIDL_user_free( void __RPC_FAR *ptr )
 {
-    HeapFree( GetProcessHeap(), 0, ptr );
+    heap_free( ptr );
 }
 
 static LONG WINAPI rpc_filter( EXCEPTION_POINTERS *eptr )
@@ -835,133 +778,6 @@
         WARN("Failed to send event, exception %#lx.\n", GetExceptionCode());
     }
     __ENDTRY
-}
-
-/***********************************************************************
- *           IoSetDeviceInterfacePropertyData   (NTOSKRNL.EXE.@)
- */
-NTSTATUS WINAPI IoSetDeviceInterfacePropertyData( UNICODE_STRING *name, const DEVPROPKEY *key, LCID lcid, ULONG flags,
-                                                  DEVPROPTYPE type, ULONG len, void *buf )
-{
-    SP_DEVICE_INTERFACE_DATA iface_data = {0};
-    WCHAR device_path[MAX_PATH];
-    struct wine_rb_entry *entry;
-    DWORD err = ERROR_SUCCESS;
-    HDEVINFO set;
-
-    TRACE( "name %s, key %s, lcid %#lx, flags %#lx, type %#lx, len %lu, buf %p\n", debugstr_us( name ),
-           debugstr_propkey( key ), lcid, flags, type, len, buf );
-
-    if (lcid != LOCALE_NEUTRAL) FIXME( "only LOCALE_NEUTRAL is supported\n" );
-
-    if (!(entry = wine_rb_get( &device_interfaces, name )))
-        return STATUS_OBJECT_NAME_NOT_FOUND;
-
-    swprintf( device_path, ARRAY_SIZE( device_path ), L"\\\\%s", &name->Buffer[2] );
-
-    if ((set = SetupDiCreateDeviceInfoListExW( NULL, NULL, NULL, NULL )) == INVALID_HANDLE_VALUE)
-    {
-        err = GetLastError();
-        ERR( "Failed to create device list, error %lu.\n", err );
-        goto done;
-    }
-    iface_data.cbSize = sizeof( iface_data );
-    if (!SetupDiOpenDeviceInterfaceW( set, device_path, 0, &iface_data ))
-    {
-        err = GetLastError();
-        ERR( "Failed to open device interface, error %lu.\n", err );
-        goto done;
-    }
-    if (!SetupDiSetDeviceInterfacePropertyW( set, &iface_data, key, buf ? type : DEVPROP_TYPE_EMPTY, (BYTE *)buf, len, 0 ))
-        err = GetLastError();
-
-done:
-    if (set != INVALID_HANDLE_VALUE)
-    {
-        SetupDiDeleteDeviceInterfaceData( set, &iface_data );
-        SetupDiDestroyDeviceInfoList( set );
-    }
-    switch (err)
-    {
-    case ERROR_SUCCESS:
-        return STATUS_SUCCESS;
-    case ERROR_INVALID_PARAMETER:
-    case ERROR_INVALID_DATA:
-    case ERROR_INVALID_FLAGS:
-        return STATUS_INVALID_PARAMETER;
-    case ERROR_NOT_ENOUGH_MEMORY:
-        return STATUS_NO_MEMORY;
-    case ERROR_NO_SUCH_DEVICE_INTERFACE:
-        return STATUS_OBJECT_NAME_NOT_FOUND;
-    case ERROR_INSUFFICIENT_BUFFER:
-        return STATUS_BUFFER_TOO_SMALL;
-    case ERROR_INVALID_ACCESS:
-        return STATUS_ACCESS_DENIED;
-    default:
-        FIXME( "Unhandled error: %lu\n", err );
-        return STATUS_INTERNAL_ERROR;
-    }
-}
-
-/***********************************************************************
- *           IoGetDeviceInterfacePropertyData   (NTOSKRNL.EXE.@)
- */
-NTSTATUS WINAPI IoGetDeviceInterfacePropertyData( UNICODE_STRING *name, const DEVPROPKEY *key, LCID lcid, ULONG flags,
-                                                  ULONG size, void *buf, ULONG *required, DEVPROPTYPE *type )
-{
-    SP_DEVICE_INTERFACE_DATA iface_data = {0};
-    WCHAR device_path[MAX_PATH];
-    struct wine_rb_entry *entry;
-    DWORD err = ERROR_SUCCESS;
-    HDEVINFO set;
-
-    TRACE( "name %s, key %s, lcid %#lx, flags %#lx, size %lu, buf %p, required %p, type %p\n",
-           debugstr_us( name ), debugstr_propkey( key ), lcid, flags, size, buf, required, type );
-
-    if (!(entry = wine_rb_get( &device_interfaces, name )))
-        return STATUS_OBJECT_NAME_NOT_FOUND;
-
-    swprintf( device_path, ARRAY_SIZE( device_path ), L"\\\\%s", &name->Buffer[2] );
-
-    if ((set = SetupDiCreateDeviceInfoListExW( NULL, NULL, NULL, NULL )) == INVALID_HANDLE_VALUE)
-    {
-        err = GetLastError();
-        ERR( "Failed to create device list, error %lu.\n", err );
-        goto done;
-    }
-    iface_data.cbSize = sizeof( iface_data );
-    if (!SetupDiOpenDeviceInterfaceW( set, device_path, 0, &iface_data ))
-    {
-        err = GetLastError();
-        ERR( "Failed to open device interface, error %lu.\n", err );
-        goto done;
-    }
-    if (!SetupDiGetDeviceInterfacePropertyW( set, &iface_data, key, type, (BYTE *)buf, size, required, flags ))
-        err = GetLastError();
-
-done:
-    if (set != INVALID_HANDLE_VALUE)
-    {
-        SetupDiDeleteDeviceInterfaceData( set, &iface_data );
-        SetupDiDestroyDeviceInfoList( set );
-    }
-    switch (err)
-    {
-    case ERROR_SUCCESS:
-        return STATUS_SUCCESS;
-    case ERROR_INVALID_PARAMETER:
-    case ERROR_INVALID_FLAGS:
-        return STATUS_INVALID_PARAMETER;
-    case ERROR_NOT_ENOUGH_MEMORY:
-        return STATUS_NO_MEMORY;
-    case ERROR_NO_SUCH_DEVICE_INTERFACE:
-        return STATUS_OBJECT_NAME_NOT_FOUND;
-    case ERROR_INSUFFICIENT_BUFFER:
-        return STATUS_BUFFER_TOO_SMALL;
-    default:
-        FIXME( "Unhandled error: %lu\n", err );
-        return STATUS_INTERNAL_ERROR;
-    }
 }
 
 /***********************************************************************
@@ -1011,7 +827,7 @@
 
     len = lstrlenW(DeviceClassesW) + 38 + 1 + namelen + 2 + 1;
 
-    if (!(path = malloc( len * sizeof(WCHAR) )))
+    if (!(path = heap_alloc( len * sizeof(WCHAR) )))
         return STATUS_NO_MEMORY;
 
     lstrcpyW( path, DeviceClassesW );
@@ -1029,7 +845,7 @@
     attr.ObjectName = &string;
     RtlInitUnicodeString( &string, path );
     ret = NtOpenKey( &iface_key, KEY_CREATE_SUB_KEY, &attr );
-    free(path);
+    heap_free(path);
     if (ret)
         return ret;
 
@@ -1063,7 +879,7 @@
 
     len = offsetof(DEV_BROADCAST_DEVICEINTERFACE_W, dbcc_name[namelen + 1]);
 
-    if ((broadcast = malloc( len )))
+    if ((broadcast = heap_alloc( len )))
     {
         broadcast->dbcc_size       = len;
         broadcast->dbcc_devicetype = DBT_DEVTYP_DEVICEINTERFACE;
@@ -1072,7 +888,7 @@
         lstrcpynW( broadcast->dbcc_name, name->Buffer, namelen + 1 );
         if (namelen > 1) broadcast->dbcc_name[1] = '\\';
         send_devicechange( L"", enable ? DBT_DEVICEARRIVAL : DBT_DEVICEREMOVECOMPLETE, broadcast, len );
-        free( broadcast );
+        heap_free( broadcast );
     }
     return ret;
 }
@@ -1192,7 +1008,7 @@
     }
     else
     {
-        iface = calloc( 1, sizeof(struct device_interface) );
+        iface = heap_alloc_zero( sizeof(struct device_interface) );
         RtlDuplicateUnicodeString( 1, &device_path, &iface->symbolic_link );
         if (wine_rb_put( &device_interfaces, &iface->symbolic_link, &iface->entry ))
             ERR("Failed to insert interface %s into tree.\n", debugstr_us(&iface->symbolic_link));
@@ -1225,15 +1041,15 @@
 
     ret = ObQueryNameString( device, NULL, 0, &size );
     if (ret != STATUS_INFO_LENGTH_MISMATCH) return ret;
-    if (!(name_info = malloc( size ))) return STATUS_NO_MEMORY;
+    if (!(name_info = heap_alloc( size ))) return STATUS_NO_MEMORY;
     ret = ObQueryNameString( device, name_info, size, &size );
     if (ret != STATUS_SUCCESS) return ret;
 
     data_size = notification->Size - offsetof( TARGET_DEVICE_CUSTOM_NOTIFICATION, CustomDataBuffer );
     size = offsetof( DEV_BROADCAST_HANDLE, dbch_data[data_size + 2 * sizeof(WCHAR)] );
-    if (!(event_handle = calloc( 1, size )))
-    {
-        free( name_info );
+    if (!(event_handle = heap_alloc_zero( size )))
+    {
+        heap_free( name_info );
         return STATUS_NO_MEMORY;
     }
 
@@ -1243,8 +1059,8 @@
     event_handle->dbch_nameoffset = notification->NameBufferOffset;
     memcpy( event_handle->dbch_data, notification->CustomDataBuffer, data_size );
     send_devicechange( name_info->Name.Buffer, DBT_CUSTOMEVENT, (BYTE *)event_handle, event_handle->dbch_size );
-    free( event_handle );
-    free( name_info );
+    heap_free( event_handle );
+    heap_free( name_info );
 
     return STATUS_SUCCESS;
 }
@@ -1435,30 +1251,6 @@
     return STATUS_SUCCESS;
 }
 
-static DWORD CALLBACK device_enum_thread_proc(void *arg)
-{
-    for (;;)
-    {
-        DEVICE_OBJECT *device;
-
-        EnterCriticalSection( &invalidated_devices_cs );
-
-        while (!invalidated_devices_count)
-            SleepConditionVariableCS( &invalidated_devices_cv, &invalidated_devices_cs, INFINITE );
-
-        device = invalidated_devices[--invalidated_devices_count];
-
-        /* Don't hold the CS while enumerating the device. Tests show that
-         * calling IoInvalidateDeviceRelations() from another thread shouldn't
-         * block, even if this thread is blocked in an IRP handler. */
-        LeaveCriticalSection( &invalidated_devices_cs );
-
-        handle_bus_relations( device );
-    }
-
-    return 0;
-}
-
 void pnp_manager_start(void)
 {
     WCHAR endpoint[] = L"\\pipe\\wine_plugplay";
@@ -1480,8 +1272,6 @@
     RpcStringFreeW( &binding_str );
     if (err)
         ERR("RpcBindingFromStringBinding() failed, error %#lx\n", err);
-
-    CreateThread( NULL, 0, device_enum_thread_proc, NULL, 0, NULL );
 }
 
 void pnp_manager_stop_driver( struct wine_driver *driver )
