diff --git a/dlls/sxs/cache.c b/dlls/sxs/cache.c
--- a/dlls/sxs/cache.c
+++ b/dlls/sxs/cache.c
@@ -98,64 +98,29 @@
     return len + ARRAY_SIZE(winsxsW) - 1;
 }
 
-static void append_string( WCHAR *buffer, const WCHAR *str, unsigned int maxlen )
-{
-    static const WCHAR valid_chars[] = L"-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
-    unsigned int len;
-
-    if (!*str) return;
-    buffer += wcslen( buffer );
-    for (len = 0; *str; str++) if (wcschr( valid_chars, *str )) buffer[len++] = *str;
-    if (len > maxlen)
-    {
-        unsigned int pos = maxlen / 2;
-        buffer[pos - 1] = buffer[pos] = '.';
-        memmove( buffer + pos + 1, buffer + len - pos + 1, (pos - 1) * sizeof(WCHAR) );
-        len = maxlen;
-    }
-    buffer[len++] = '_';
-    buffer[len] = 0;
-    wcslwr( buffer );
-}
-
 static WCHAR *build_assembly_name( const WCHAR *arch, const WCHAR *name, const WCHAR *token,
-                                   const WCHAR *version, const WCHAR *language, unsigned int *len )
-{
-    unsigned int buflen = 20;
+                                   const WCHAR *version, unsigned int *len )
+{
+    static const WCHAR fmtW[] = L"%s_%s_%s_%s_none_deadbeef";
+    unsigned int buflen = ARRAY_SIZE(fmtW);
     WCHAR *ret;
 
-    if (!arch) arch = L"none";
-    if (!token) token = L"none";
-    if (!language) language = L"none";
-
-    buflen += wcslen( arch ) + 1;
-    buflen += wcslen( name ) + 1;
-    buflen += wcslen( token ) + 1;
-    buflen += wcslen( version ) + 1;
-    buflen += wcslen( language ) + 1;
+    buflen += lstrlenW( arch );
+    buflen += lstrlenW( name );
+    buflen += lstrlenW( token );
+    buflen += lstrlenW( version );
     if (!(ret = malloc( buflen * sizeof(WCHAR) ))) return NULL;
-    ret[0] = 0;
-    append_string( ret, arch, 16 );
-    append_string( ret, name, 40 );
-    wcscat( ret, token );
-    wcscat( ret, L"_" );
-    wcscat( ret, version );
-    wcscat( ret, L"_" );
-    append_string( ret, language, 8 );
-    wcscat( ret, L"deadbeef" );
-    *len = wcslen( ret );
-    return ret;
+    *len = swprintf( ret, buflen, fmtW, arch, name, token, version );
+    return wcslwr( ret );
 }
 
 static WCHAR *build_dll_path( const WCHAR *arch, const WCHAR *name, const WCHAR *token,
-                              const WCHAR *version, const WCHAR *language )
+                              const WCHAR *version )
 {
     WCHAR *path = NULL, *ret, sxsdir[MAX_PATH];
     unsigned int len;
 
-    if (!language || !wcsicmp( language, L"neutral" ) || !wcscmp( language, L"*")) language = L"none";
-
-    if (!(path = build_assembly_name( arch, name, token, version, language, &len ))) return NULL;
+    if (!(path = build_assembly_name( arch, name, token, version, &len ))) return NULL;
     len += build_sxs_path( sxsdir ) + 2;
     if (!(ret = malloc( len * sizeof(WCHAR) )))
     {
@@ -169,6 +134,42 @@
     return ret;
 }
 
+static WCHAR *build_policy_name( const WCHAR *arch, const WCHAR *name, const WCHAR *token,
+                                 unsigned int *len )
+{
+    static const WCHAR fmtW[] = L"%s_%s_%s_none_deadbeef";
+    unsigned int buflen = ARRAY_SIZE(fmtW);
+    WCHAR *ret;
+
+    buflen += lstrlenW( arch );
+    buflen += lstrlenW( name );
+    buflen += lstrlenW( token );
+    if (!(ret = malloc( buflen * sizeof(WCHAR) ))) return NULL;
+    *len = swprintf( ret, buflen, fmtW, arch, name, token );
+    return wcslwr( ret );
+}
+
+static WCHAR *build_policy_path( const WCHAR *arch, const WCHAR *name, const WCHAR *token,
+                                 const WCHAR *version )
+{
+    static const WCHAR fmtW[] = L"%spolicies\\%s\\%s.policy";
+    WCHAR *path = NULL, *ret, sxsdir[MAX_PATH];
+    unsigned int len;
+
+    if (!(path = build_policy_name( arch, name, token, &len ))) return NULL;
+    len += ARRAY_SIZE(fmtW);
+    len += build_sxs_path( sxsdir );
+    len += lstrlenW( version );
+    if (!(ret = malloc( len * sizeof(WCHAR) )))
+    {
+        free( path );
+        return NULL;
+    }
+    swprintf( ret, len, fmtW, sxsdir, path, version );
+    free( path );
+    return ret;
+}
+
 static void cache_lock( struct cache *cache )
 {
     WaitForSingleObject( cache->lock, INFINITE );
@@ -189,7 +190,7 @@
 {
     struct cache *cache = impl_from_IAssemblyCache( iface );
     IAssemblyName *name_obj;
-    const WCHAR *arch, *name, *token, *type, *version, *language;
+    const WCHAR *arch, *name, *token, *type, *version;
     WCHAR *path = NULL;
     unsigned int len;
     HRESULT hr;
@@ -208,7 +209,6 @@
     token = get_name_attribute( name_obj, NAME_ATTR_ID_TOKEN );
     type = get_name_attribute( name_obj, NAME_ATTR_ID_TYPE );
     version = get_name_attribute( name_obj, NAME_ATTR_ID_VERSION );
-    language = get_name_attribute( name_obj, NAME_ATTR_ID_LANGUAGE );
     if (!arch || !name || !token || !type || !version)
     {
         IAssemblyName_Release( name_obj );
@@ -219,18 +219,20 @@
         IAssemblyName_Release( name_obj );
         return S_OK;
     }
-    if (wcscmp( type, L"win32" ) && wcscmp( type, L"win32-policy" ))
+    cache_lock( cache );
+
+    if (!wcscmp( type, L"win32" )) path = build_dll_path( arch, name, token, version );
+    else if (!wcscmp( type, L"win32-policy" )) path = build_policy_path( arch, name, token, version );
+    else
     {
         hr = HRESULT_FROM_WIN32( ERROR_SXS_INVALID_IDENTITY_ATTRIBUTE_VALUE );
         goto done;
     }
-    if (!(path = build_dll_path( arch, name, token, version, language )))
+    if (!path)
     {
         hr = E_OUTOFMEMORY;
         goto done;
     }
-
-    cache_lock( cache );
     hr = S_OK;
     if (GetFileAttributesW( path ) != INVALID_FILE_ATTRIBUTES) /* FIXME: better check */
     {
@@ -247,11 +249,11 @@
         }
         else lstrcpyW( info->pszCurrentAssemblyPathBuf, path );
     }
-    cache_unlock( cache );
 
 done:
     free( path );
     IAssemblyName_Release( name_obj );
+    cache_unlock( cache );
     return hr;
 }
 
@@ -311,7 +313,6 @@
     BSTR version;
     BSTR arch;
     BSTR token;
-    BSTR language;
     struct list files;
 };
 
@@ -325,7 +326,6 @@
     SysFreeString( assembly->version );
     SysFreeString( assembly->arch );
     SysFreeString( assembly->token );
-    SysFreeString( assembly->language );
     LIST_FOR_EACH_SAFE( item, cursor, &assembly->files )
     {
         struct file *file = LIST_ENTRY( item, struct file, entry );
@@ -449,7 +449,6 @@
     a->version = get_attribute_value( attrs, L"version" );
     a->arch    = get_attribute_value( attrs, L"processorArchitecture" );
     a->token   = get_attribute_value( attrs, L"publicKeyToken" );
-    a->language = get_attribute_value( attrs, L"language" );
 
     if (!a->type || (wcscmp( a->type, L"win32" ) && wcscmp( a->type, L"win32-policy" )) ||
         !a->name || !a->version || !a->arch || !a->token)
@@ -469,6 +468,58 @@
     return hr;
 }
 
+static WCHAR *build_policy_filename( const WCHAR *arch, const WCHAR *name, const WCHAR *token,
+                                     const WCHAR *version )
+{
+    static const WCHAR policiesW[] = L"policies\\";
+    static const WCHAR suffixW[] = L".policy";
+    WCHAR sxsdir[MAX_PATH], *ret, *fullname;
+    unsigned int len;
+
+    if (!(fullname = build_policy_name( arch, name, token, &len ))) return NULL;
+    len += build_sxs_path( sxsdir );
+    len += ARRAY_SIZE(policiesW) - 1;
+    len += lstrlenW( version );
+    len += ARRAY_SIZE(suffixW) - 1;
+    if (!(ret = malloc( (len + 1) * sizeof(WCHAR) )))
+    {
+        free( fullname );
+        return NULL;
+    }
+    lstrcpyW( ret, sxsdir );
+    lstrcatW( ret, policiesW );
+    CreateDirectoryW( ret, NULL );
+    lstrcatW( ret, name );
+    CreateDirectoryW( ret, NULL );
+    lstrcatW( ret, L"\\" );
+    lstrcatW( ret, version );
+    lstrcatW( ret, suffixW );
+
+    free( fullname );
+    return ret;
+}
+
+static HRESULT install_policy( const WCHAR *manifest, struct assembly *assembly )
+{
+    WCHAR *dst;
+    BOOL ret;
+
+    /* FIXME: handle catalog file */
+
+    dst = build_policy_filename( assembly->arch, assembly->name, assembly->token, assembly->version );
+    if (!dst) return E_OUTOFMEMORY;
+
+    ret = CopyFileW( manifest, dst, FALSE );
+    free( dst );
+    if (!ret)
+    {
+        HRESULT hr = HRESULT_FROM_WIN32( GetLastError() );
+        WARN("failed to copy policy manifest file 0x%08lx\n", hr);
+        return hr;
+    }
+    return S_OK;
+}
+
 static WCHAR *build_source_filename( const WCHAR *manifest, struct file *file )
 {
     WCHAR *src;
@@ -489,14 +540,14 @@
 }
 
 static WCHAR *build_manifest_filename( const WCHAR *arch, const WCHAR *name, const WCHAR *token,
-                                       const WCHAR *version, const WCHAR *language )
+                                       const WCHAR *version )
 {
     static const WCHAR manifestsW[] = L"manifests\\";
     static const WCHAR suffixW[] = L".manifest";
     WCHAR sxsdir[MAX_PATH], *ret, *fullname;
     unsigned int len;
 
-    if (!(fullname = build_assembly_name( arch, name, token, version, language, &len ))) return NULL;
+    if (!(fullname = build_assembly_name( arch, name, token, version, &len ))) return NULL;
     len += build_sxs_path( sxsdir );
     len += ARRAY_SIZE(manifestsW) - 1;
     len += ARRAY_SIZE(suffixW) - 1;
@@ -544,8 +595,7 @@
     HRESULT hr = E_OUTOFMEMORY;
     BOOL ret;
 
-    dst = build_manifest_filename( assembly->arch, assembly->name, assembly->token,
-                                   assembly->version, assembly->language );
+    dst = build_manifest_filename( assembly->arch, assembly->name, assembly->token, assembly->version );
     if (!dst) return E_OUTOFMEMORY;
 
     if (GetFileAttributesW( dst ) != INVALID_FILE_ATTRIBUTES)
@@ -565,7 +615,7 @@
     }
 
     name = build_assembly_name( assembly->arch, assembly->name, assembly->token, assembly->version,
-                                assembly->language, &len_name );
+                                &len_name );
     if (!name) return E_OUTOFMEMORY;
 
     /* FIXME: this should be a transaction */
@@ -629,7 +679,10 @@
 
     /* FIXME: verify name attributes */
 
-    hr = install_assembly( path, assembly );
+    if (!wcscmp( assembly->type, L"win32-policy" ))
+        hr = install_policy( path, assembly );
+    else
+        hr = install_assembly( path, assembly );
 
 done:
     free_assembly( assembly );
@@ -647,7 +700,7 @@
     struct file *file;
 
     name = build_assembly_name( assembly->arch, assembly->name, assembly->token, assembly->version,
-                                assembly->language, &len_name );
+                                &len_name );
     if (!name) return E_OUTOFMEMORY;
     if (!(dirname = malloc( (len_sxsdir + len_name + 1) * sizeof(WCHAR) )))
         goto done;
@@ -686,7 +739,7 @@
     IXMLDOMDocument *doc = NULL;
     struct assembly *assembly = NULL;
     IAssemblyName *name_obj = NULL;
-    const WCHAR *arch, *name, *token, *type, *version, *language;
+    const WCHAR *arch, *name, *token, *type, *version;
     WCHAR *p, *path = NULL;
 
     TRACE("%p, 0x%08lx, %s, %p, %p\n", iface, flags, debugstr_w(assembly_name), ref, disp);
@@ -708,18 +761,18 @@
     token = get_name_attribute( name_obj, NAME_ATTR_ID_TOKEN );
     type = get_name_attribute( name_obj, NAME_ATTR_ID_TYPE );
     version = get_name_attribute( name_obj, NAME_ATTR_ID_VERSION );
-    language = get_name_attribute( name_obj, NAME_ATTR_ID_LANGUAGE );
     if (!arch || !name || !token || !type || !version)
     {
         hr = E_INVALIDARG;
         goto done;
     }
-    if (wcscmp( type, L"win32" ) && wcscmp( type, L"win32-policy" ))
+    if (!wcscmp( type, L"win32" )) path = build_manifest_filename( arch, name, token, version );
+    else if (!wcscmp( type, L"win32-policy" )) path = build_policy_filename( arch, name, token, version );
+    else
     {
         hr = E_INVALIDARG;
         goto done;
     }
-    path = build_manifest_filename( arch, name, token, version, language );
 
     hr = CoCreateInstance( &CLSID_DOMDocument, NULL, CLSCTX_INPROC_SERVER, &IID_IXMLDOMDocument, (void **)&doc );
     if (hr != S_OK)
@@ -787,59 +840,3 @@
     *obj = &cache->IAssemblyCache_iface;
     return S_OK;
 }
-
-
-/******************************************************************
- *  SxspGenerateManifestPathOnAssemblyIdentity   (SXS.@)
- */
-BOOL WINAPI SxspGenerateManifestPathOnAssemblyIdentity( const WCHAR *identity, WCHAR *buffer,
-                                                        DWORD *size, void *unknown )
-{
-    const WCHAR *arch, *name, *token, *version, *language;
-    WCHAR *path = NULL;
-    BOOL ret = FALSE;
-    unsigned int retlen;
-    IAssemblyName *name_obj = NULL;
-    HRESULT hr = CreateAssemblyNameObject( &name_obj, identity, CANOF_PARSE_DISPLAY_NAME, 0 );
-
-    if (FAILED( hr ))
-    {
-        SetLastError( ERROR_INVALID_PARAMETER );
-        return FALSE;
-    }
-
-    arch = get_name_attribute( name_obj, NAME_ATTR_ID_ARCH );
-    name = get_name_attribute( name_obj, NAME_ATTR_ID_NAME );
-    token = get_name_attribute( name_obj, NAME_ATTR_ID_TOKEN );
-    version = get_name_attribute( name_obj, NAME_ATTR_ID_VERSION );
-    language = get_name_attribute( name_obj, NAME_ATTR_ID_LANGUAGE );
-    if (!name || !version)
-    {
-        SetLastError( ERROR_INVALID_PARAMETER );
-        goto done;
-    }
-    if (wcslen( version ) > 103 || (token && wcslen( token ) > 100))
-    {
-        SetLastError( ERROR_INSUFFICIENT_BUFFER );
-        goto done;
-    }
-    if (!(path = build_assembly_name( arch, name, token, version, language, &retlen )))
-    {
-        SetLastError( ERROR_OUTOFMEMORY );
-        goto done;
-    }
-    if (*size <= retlen)
-    {
-        *size = retlen + 1;
-        SetLastError( ERROR_INSUFFICIENT_BUFFER );
-    }
-    else
-    {
-        wcscpy( buffer, path );
-        ret = TRUE;
-    }
- done:
-    IAssemblyName_Release( name_obj );
-    free( path );
-    return ret;
-}
