diff --git a/dlls/comctl32/updown.c b/dlls/comctl32/updown.c
--- a/dlls/comctl32/updown.c
+++ b/dlls/comctl32/updown.c
@@ -31,6 +31,8 @@
 #include "winnls.h"
 #include "commctrl.h"
 #include "comctl32.h"
+#include "uxtheme.h"
+#include "vssym32.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(updown);
@@ -65,6 +67,8 @@
 #define DEFAULT_ADDBOT       0 /* amount to extend below the buddy window */
 #define DEFAULT_BUDDYBORDER  2 /* Width/height of the buddy border */
 #define DEFAULT_BUDDYSPACER  2 /* Spacer between the buddy and the ctrl */
+#define DEFAULT_BUDDYBORDER_THEMED  1 /* buddy border when theming is enabled */
+#define DEFAULT_BUDDYSPACER_THEMED  0 /* buddy spacer when theming is enabled */
 
 /* Work constants */
 
@@ -159,24 +163,6 @@
 	      UPDOWN_IsBuddyEdit(infoPtr) );
 }
 
-static int UPDOWN_GetBuddyBorderSize(HWND hwnd)
-{
-#if __WINE_COMCTL32_VERSION == 6
-    if (GetWindowTheme(hwnd)) return 1;
-#endif
-
-    return DEFAULT_BUDDYBORDER;
-}
-
-static int UPDOWN_GetBuddySpacerSize(HWND hwnd)
-{
-#if __WINE_COMCTL32_VERSION == 6
-    if (GetWindowTheme(hwnd)) return 0;
-#endif
-
-    return DEFAULT_BUDDYSPACER;
-}
-
 /***********************************************************************
  *           UPDOWN_GetArrowRect
  * wndPtr   - pointer to the up-down wnd
@@ -186,8 +172,9 @@
  */
 static void UPDOWN_GetArrowRect (const UPDOWN_INFO* infoPtr, RECT *rect, unsigned int arrow)
 {
-    const int border = UPDOWN_GetBuddyBorderSize(infoPtr->Self);
-    const int spacer = UPDOWN_GetBuddySpacerSize(infoPtr->Self);
+    HTHEME theme = GetWindowTheme (infoPtr->Self);
+    const int border = theme ? DEFAULT_BUDDYBORDER_THEMED : DEFAULT_BUDDYBORDER;
+    const int spacer = theme ? DEFAULT_BUDDYSPACER_THEMED : DEFAULT_BUDDYSPACER;
     int size;
 
     assert(arrow && (arrow & (FLAG_INCR | FLAG_DECR)) != (FLAG_INCR | FLAG_DECR));
@@ -366,165 +353,87 @@
  *
  * Draw buddy background for visual integration.
  */
-static void UPDOWN_DrawBuddyBackground (const UPDOWN_INFO *infoPtr, HDC hdc)
-{
+static BOOL UPDOWN_DrawBuddyBackground (const UPDOWN_INFO *infoPtr, HDC hdc)
+{
+    RECT br, r;
+    HTHEME buddyTheme = GetWindowTheme (infoPtr->Buddy);
+    if (!buddyTheme) return FALSE;
+
+    GetWindowRect (infoPtr->Buddy, &br);
+    MapWindowPoints (NULL, infoPtr->Self, (POINT*)&br, 2);
+    GetClientRect (infoPtr->Self, &r);
+
+    if (infoPtr->dwStyle & UDS_ALIGNLEFT)
+        br.left = r.left;
+    else if (infoPtr->dwStyle & UDS_ALIGNRIGHT)
+        br.right = r.right;
+    /* FIXME: take disabled etc. into account */
+    DrawThemeBackground (buddyTheme, hdc, 0, 0, &br, NULL);
+    return TRUE;
+}
+
+/***********************************************************************
+ * UPDOWN_Draw
+ *
+ * Draw the arrows. The background need not be erased.
+ */
+static LRESULT UPDOWN_Draw (const UPDOWN_INFO *infoPtr, HDC hdc)
+{
+    BOOL uPressed, uHot, dPressed, dHot;
     RECT rect;
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme(infoPtr->Self);
-    HTHEME buddyTheme = GetWindowTheme (infoPtr->Buddy);
-
-    if (theme && buddyTheme)
-    {
-        RECT br;
-
-        GetWindowRect(infoPtr->Buddy, &br);
-        MapWindowPoints(NULL, infoPtr->Self, (POINT*)&br, 2);
-        GetClientRect(infoPtr->Self, &rect);
-
-        if (infoPtr->dwStyle & UDS_ALIGNLEFT)
-            br.left = rect.left;
-        else if (infoPtr->dwStyle & UDS_ALIGNRIGHT)
-            br.right = rect.right;
-        /* FIXME: take disabled etc. into account */
-        DrawThemeBackground(buddyTheme, hdc, 0, 0, &br, NULL);
-        return;
-    }
-#endif
-
-    GetClientRect(infoPtr->Self, &rect);
-    DrawEdge(hdc, &rect, EDGE_SUNKEN, BF_BOTTOM | BF_TOP | (infoPtr->dwStyle & UDS_ALIGNLEFT ? BF_LEFT : BF_RIGHT));
-}
-
-static BOOL UPDOWN_IsUpArrowPressed(const UPDOWN_INFO *infoPtr)
-{
-    return infoPtr->Flags & FLAG_PRESSED && infoPtr->Flags & FLAG_INCR;
-}
-
-static BOOL UPDOWN_IsUpArrowHot(const UPDOWN_INFO *infoPtr)
-{
-    return infoPtr->Flags & FLAG_INCR && infoPtr->Flags & FLAG_MOUSEIN;
-}
-
-static BOOL UPDOWN_IsDownArrowPressed(const UPDOWN_INFO *infoPtr)
-{
-    return infoPtr->Flags & FLAG_PRESSED && infoPtr->Flags & FLAG_DECR;
-}
-
-static BOOL UPDOWN_IsDownArrowHot(const UPDOWN_INFO *infoPtr)
-{
-    return infoPtr->Flags & FLAG_DECR && infoPtr->Flags & FLAG_MOUSEIN;
-}
-
-#if __WINE_COMCTL32_VERSION == 6
-static void UPDOWN_GetUpArrowThemePartAndState(const UPDOWN_INFO *infoPtr, int *part, int *state)
-{
-    BOOL pressed, hot;
-
-    pressed = UPDOWN_IsUpArrowPressed(infoPtr);
-    hot = UPDOWN_IsUpArrowHot(infoPtr);
-    *part = (infoPtr->dwStyle & UDS_HORZ) ? SPNP_UPHORZ : SPNP_UP;
-    *state = (infoPtr->dwStyle & WS_DISABLED) ? DNS_DISABLED : (pressed ? DNS_PRESSED : (hot ? DNS_HOT : DNS_NORMAL));
-}
-
-static void UPDOWN_GetDownArrowThemePartAndState(const UPDOWN_INFO *infoPtr, int *part, int *state)
-{
-    BOOL pressed, hot;
-
-    pressed = UPDOWN_IsDownArrowPressed(infoPtr);
-    hot = UPDOWN_IsDownArrowHot(infoPtr);
-    *part = (infoPtr->dwStyle & UDS_HORZ) ? SPNP_DOWNHORZ : SPNP_DOWN;
-    *state = (infoPtr->dwStyle & WS_DISABLED) ? DNS_DISABLED : (pressed ? DNS_PRESSED : (hot ? DNS_HOT : DNS_NORMAL));
-}
-#endif /* __WINE_COMCTL32_VERSION == 6 */
-
-static BOOL UPDOWN_NeedBuddyBackground(const UPDOWN_INFO *infoPtr)
-{
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme(infoPtr->Self);
-
-    if (theme)
-    {
-        int up_part = 0, up_state = 0, down_part = 0, down_state = 0;
-        BOOL up_transparent, down_transparent, need_buddy_bg;
-
-        UPDOWN_GetUpArrowThemePartAndState(infoPtr, &up_part, &up_state);
-        UPDOWN_GetDownArrowThemePartAndState(infoPtr, &down_part, &down_state);
-        up_transparent = IsThemeBackgroundPartiallyTransparent(theme, up_part, up_state);
-        down_transparent = IsThemeBackgroundPartiallyTransparent(theme, down_part, down_state);
-        need_buddy_bg = IsWindow(infoPtr->Buddy) && (up_transparent || down_transparent);
-        return UPDOWN_HasBuddyBorder(infoPtr) || need_buddy_bg;
-    }
-#endif
-
-    return UPDOWN_HasBuddyBorder(infoPtr);
-}
-
-static void UPDOWN_DrawUpArrow(const UPDOWN_INFO *infoPtr, HDC hdc)
-{
-    RECT rect;
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme(infoPtr->Self);
-
-    if (theme)
-    {
-        int part = 0, state = 0;
-
-        UPDOWN_GetUpArrowThemePartAndState(infoPtr, &part, &state);
-        UPDOWN_GetArrowRect(infoPtr, &rect, FLAG_INCR);
-        DrawThemeBackground(theme, hdc, part, state, &rect, NULL);
-        return;
-    }
-#endif
-
-    UPDOWN_GetArrowRect(infoPtr, &rect, FLAG_INCR);
-    DrawFrameControl(hdc, &rect, DFC_SCROLL,
-                     (infoPtr->dwStyle & UDS_HORZ ? DFCS_SCROLLRIGHT : DFCS_SCROLLUP) |
-                     ((infoPtr->dwStyle & UDS_HOTTRACK) && UPDOWN_IsUpArrowHot(infoPtr) ? DFCS_HOT : 0) |
-                     (UPDOWN_IsUpArrowPressed(infoPtr) ? DFCS_PUSHED : 0) |
-                     (infoPtr->dwStyle & WS_DISABLED ? DFCS_INACTIVE : 0));
-}
-
-static void UPDOWN_DrawDownArrow(const UPDOWN_INFO *infoPtr, HDC hdc)
-{
-    RECT rect;
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme(infoPtr->Self);
-
-    if (theme)
-    {
-        int part = 0, state = 0;
-
-        UPDOWN_GetDownArrowThemePartAndState(infoPtr, &part, &state);
-        UPDOWN_GetArrowRect(infoPtr, &rect, FLAG_DECR);
-        DrawThemeBackground(theme, hdc, part, state, &rect, NULL);
-        return;
-    }
-#endif
-
+    HTHEME theme = GetWindowTheme (infoPtr->Self);
+    int uPart = 0, uState = 0, dPart = 0, dState = 0;
+    BOOL needBuddyBg = FALSE;
+
+    uPressed = (infoPtr->Flags & FLAG_PRESSED) && (infoPtr->Flags & FLAG_INCR);
+    uHot = (infoPtr->Flags & FLAG_INCR) && (infoPtr->Flags & FLAG_MOUSEIN);
+    dPressed = (infoPtr->Flags & FLAG_PRESSED) && (infoPtr->Flags & FLAG_DECR);
+    dHot = (infoPtr->Flags & FLAG_DECR) && (infoPtr->Flags & FLAG_MOUSEIN);
+    if (theme) {
+        uPart = (infoPtr->dwStyle & UDS_HORZ) ? SPNP_UPHORZ : SPNP_UP;
+        uState = (infoPtr->dwStyle & WS_DISABLED) ? DNS_DISABLED 
+            : (uPressed ? DNS_PRESSED : (uHot ? DNS_HOT : DNS_NORMAL));
+        dPart = (infoPtr->dwStyle & UDS_HORZ) ? SPNP_DOWNHORZ : SPNP_DOWN;
+        dState = (infoPtr->dwStyle & WS_DISABLED) ? DNS_DISABLED 
+            : (dPressed ? DNS_PRESSED : (dHot ? DNS_HOT : DNS_NORMAL));
+        needBuddyBg = IsWindow (infoPtr->Buddy)
+            && (IsThemeBackgroundPartiallyTransparent (theme, uPart, uState)
+              || IsThemeBackgroundPartiallyTransparent (theme, dPart, dState));
+    }
+
+    /* Draw the common border between ourselves and our buddy */
+    if (UPDOWN_HasBuddyBorder(infoPtr) || needBuddyBg) {
+        if (!theme || !UPDOWN_DrawBuddyBackground (infoPtr, hdc)) {
+            GetClientRect(infoPtr->Self, &rect);
+	    DrawEdge(hdc, &rect, EDGE_SUNKEN,
+		     BF_BOTTOM | BF_TOP |
+		     (infoPtr->dwStyle & UDS_ALIGNLEFT ? BF_LEFT : BF_RIGHT));
+        }
+    }
+
+    /* Draw the incr button */
+    UPDOWN_GetArrowRect (infoPtr, &rect, FLAG_INCR);
+    if (theme) {
+        DrawThemeBackground(theme, hdc, uPart, uState, &rect, NULL);
+    } else {
+        DrawFrameControl(hdc, &rect, DFC_SCROLL,
+            (infoPtr->dwStyle & UDS_HORZ ? DFCS_SCROLLRIGHT : DFCS_SCROLLUP) |
+            ((infoPtr->dwStyle & UDS_HOTTRACK) && uHot ? DFCS_HOT : 0) |
+            (uPressed ? DFCS_PUSHED : 0) |
+            (infoPtr->dwStyle & WS_DISABLED ? DFCS_INACTIVE : 0) );
+    }
+
+    /* Draw the decr button */
     UPDOWN_GetArrowRect(infoPtr, &rect, FLAG_DECR);
-    DrawFrameControl(hdc, &rect, DFC_SCROLL,
-                     (infoPtr->dwStyle & UDS_HORZ ? DFCS_SCROLLLEFT : DFCS_SCROLLDOWN) |
-                     ((infoPtr->dwStyle & UDS_HOTTRACK) && UPDOWN_IsDownArrowHot(infoPtr) ? DFCS_HOT : 0) |
-                     (UPDOWN_IsDownArrowPressed(infoPtr) ? DFCS_PUSHED : 0) |
-                     (infoPtr->dwStyle & WS_DISABLED ? DFCS_INACTIVE : 0) );
-}
-
-/***********************************************************************
- * UPDOWN_Draw
- *
- * Draw the arrows. The background need not be erased.
- */
-static LRESULT UPDOWN_Draw (const UPDOWN_INFO *infoPtr, HDC hdc)
-{
-    /* Draw the common border between ourselves and our buddy */
-    if (UPDOWN_NeedBuddyBackground(infoPtr))
-        UPDOWN_DrawBuddyBackground(infoPtr, hdc);
-
-    /* Draw the incr button */
-    UPDOWN_DrawUpArrow(infoPtr, hdc);
-
-    /* Draw the decr button */
-    UPDOWN_DrawDownArrow(infoPtr, hdc);
+    if (theme) {
+        DrawThemeBackground(theme, hdc, dPart, dState, &rect, NULL);
+    } else {
+        DrawFrameControl(hdc, &rect, DFC_SCROLL,
+            (infoPtr->dwStyle & UDS_HORZ ? DFCS_SCROLLLEFT : DFCS_SCROLLDOWN) |
+            ((infoPtr->dwStyle & UDS_HOTTRACK) && dHot ? DFCS_HOT : 0) |
+            (dPressed ? DFCS_PUSHED : 0) |
+            (infoPtr->dwStyle & WS_DISABLED ? DFCS_INACTIVE : 0) );
+    }
 
     return 0;
 }
@@ -609,7 +518,6 @@
         else
             ret = infoPtr->MaxVal;
     }
-    NotifyWinEvent( EVENT_OBJECT_VALUECHANGE, infoPtr->Self, OBJID_CLIENT, 0 );
     return ret;
 }
 
@@ -980,6 +888,7 @@
 static LRESULT WINAPI UpDownWindowProc(HWND hwnd, UINT message, WPARAM wParam, LPARAM lParam)
 {
     UPDOWN_INFO *infoPtr = UPDOWN_GetInfoPtr (hwnd);
+    HTHEME theme;
 
     TRACE("hwnd %p, msg %04x, wparam %Id, lparam %Ix\n", hwnd, message, wParam, lParam);
 
@@ -999,6 +908,9 @@
 	    infoPtr->Self = hwnd;
 	    infoPtr->Notify  = pcs->hwndParent;
 	    infoPtr->dwStyle = pcs->style;
+	    infoPtr->AccelCount = 0;
+	    infoPtr->AccelVect = 0;
+	    infoPtr->AccelIndex = -1;
 	    infoPtr->CurVal = 0;
 	    infoPtr->MinVal = 100;
 	    infoPtr->MaxVal = 0;
@@ -1006,17 +918,6 @@
 	    infoPtr->Buddy = 0;  /* No buddy window yet */
 	    infoPtr->Flags = (infoPtr->dwStyle & UDS_SETBUDDYINT) ? FLAG_BUDDYINT : 0;
 
-            infoPtr->AccelCount = 3;
-            infoPtr->AccelIndex = -1;
-            infoPtr->AccelVect = Alloc(infoPtr->AccelCount * sizeof(UDACCEL));
-
-            infoPtr->AccelVect[0].nSec = 0;
-            infoPtr->AccelVect[0].nInc = 1;
-            infoPtr->AccelVect[1].nSec = 2;
-            infoPtr->AccelVect[1].nInc = 5;
-            infoPtr->AccelVect[2].nSec = 5;
-            infoPtr->AccelVect[2].nInc = 20;
-
             SetWindowLongW (hwnd, GWL_STYLE, infoPtr->dwStyle & ~WS_BORDER);
 	    if (!(infoPtr->dwStyle & UDS_HORZ))
 	        SetWindowPos (hwnd, NULL, 0, 0, DEFAULT_WIDTH, pcs->cy,
@@ -1026,7 +927,7 @@
 	    if (infoPtr->dwStyle & UDS_AUTOBUDDY)
 		UPDOWN_SetBuddy (infoPtr, GetWindow (hwnd, GW_HWNDPREV));
 
-            COMCTL32_OpenThemeForWindow (hwnd, L"Spin");
+	    OpenThemeData (hwnd, L"Spin");
 
 	    TRACE("UpDown Ctrl creation, hwnd=%p\n", hwnd);
 	    }
@@ -1037,8 +938,8 @@
             UPDOWN_ResetSubclass (infoPtr);
 	    Free (infoPtr);
 	    SetWindowLongPtrW (hwnd, 0, 0);
-            COMCTL32_CloseThemeForWindow (hwnd);
-
+            theme = GetWindowTheme (hwnd);
+            CloseThemeData (theme);
 	    TRACE("UpDown Ctrl destruction, hwnd=%p\n", hwnd);
 	    break;
 
@@ -1052,18 +953,17 @@
 	    InvalidateRect (infoPtr->Self, NULL, FALSE);
 	    break;
 
-        case WM_GETOBJECT:
-            if ((LONG)lParam == OBJID_QUERYCLASSNAMEIDX)
-                return 0x10016;
-	    return DefWindowProcW (hwnd, message, wParam, lParam);
-
         case WM_STYLECHANGED:
             if (wParam == GWL_STYLE)
                 infoPtr->dwStyle = ((LPSTYLESTRUCT)lParam)->styleNew;
             break;
 
         case WM_THEMECHANGED:
-            return COMCTL32_ThemeChanged (hwnd, L"Spin", TRUE, TRUE);
+            theme = GetWindowTheme (hwnd);
+            CloseThemeData (theme);
+            OpenThemeData (hwnd, L"Spin");
+            InvalidateRect (hwnd, NULL, TRUE);
+            break;
 
 	case WM_TIMER:
 	   /* is this the auto-press timer? */
@@ -1145,12 +1045,14 @@
 	case WM_PAINT:
 	    return UPDOWN_Paint (infoPtr, (HDC)wParam);
 
-        case UDM_GETACCEL:
-            if (wParam && lParam) {
-                int temp = min(infoPtr->AccelCount, wParam);
-                memcpy((void *)lParam, infoPtr->AccelVect, temp * sizeof(UDACCEL));
-            }
-            return infoPtr->AccelCount;
+	case UDM_GETACCEL:
+	    if (wParam==0 && lParam==0) return infoPtr->AccelCount;
+	    if (wParam && lParam) {
+		int temp = min(infoPtr->AccelCount, wParam);
+	        memcpy((void *)lParam, infoPtr->AccelVect, temp*sizeof(UDACCEL));
+	        return temp;
+      	    }
+	    return 0;
 
 	case UDM_SETACCEL:
 	{
@@ -1280,3 +1182,14 @@
 
     RegisterClassW( &wndClass );
 }
+
+
+/***********************************************************************
+ *		UPDOWN_Unregister	[Internal]
+ *
+ * Unregisters the updown window class.
+ */
+void UPDOWN_Unregister (void)
+{
+    UnregisterClassW (UPDOWN_CLASSW, NULL);
+}
