diff --git a/dlls/msvcp90/exception.c b/dlls/msvcp90/exception.c
--- a/dlls/msvcp90/exception.c
+++ b/dlls/msvcp90/exception.c
@@ -19,27 +19,27 @@
 #include <errno.h>
 #include <stdarg.h>
 
-#define COBJMACROS
-
 #include "msvcp90.h"
 #include "windef.h"
 #include "winbase.h"
 #include "winternl.h"
 #include "rtlsupportapi.h"
-#include "unknwn.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(msvcp);
 
 #define CXX_FRAME_MAGIC_VC6 0x19930520
 
-#ifdef CXX_USE_RVA
+#ifdef RTTI_USE_RVA
 #define CXX_EXCEPTION_PARAMS 4
 #else
 #define CXX_EXCEPTION_PARAMS 3
 #endif
 
 CREATE_TYPE_INFO_VTABLE
+
+#define CLASS_IS_SIMPLE_TYPE          1
+#define CLASS_HAS_VIRTUAL_BASE_CLASS  4
 
 int* __cdecl __processing_throw(void);
 
@@ -201,11 +201,11 @@
 }
 
 #if _MSVCP_VER >= 80
-DEFINE_RTTI_DATA(exception, 0, ".?AVexception@std@@")
+DEFINE_RTTI_DATA0(exception, 0, ".?AVexception@std@@")
 #else
-DEFINE_RTTI_DATA(exception, 0, ".?AVexception@@")
-#endif
-DEFINE_CXX_TYPE(exception, MSVCP_exception_dtor)
+DEFINE_RTTI_DATA0(exception, 0, ".?AVexception@@")
+#endif
+DEFINE_CXX_DATA0(exception, MSVCP_exception_dtor)
 
 /* bad_alloc class data */
 typedef exception bad_alloc;
@@ -279,8 +279,8 @@
     return bad_alloc_copy_ctor(this, assign);
 }
 
-DEFINE_RTTI_DATA(bad_alloc, 0, ".?AVbad_alloc@std@@", exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(bad_alloc, MSVCP_bad_alloc_dtor, exception_cxx_type_info)
+DEFINE_RTTI_DATA1(bad_alloc, 0, &exception_rtti_base_descriptor, ".?AVbad_alloc@std@@")
+DEFINE_CXX_DATA1(bad_alloc, &exception_cxx_type_info, MSVCP_bad_alloc_dtor)
 
 /* logic_error class data */
 typedef struct {
@@ -407,11 +407,11 @@
 }
 
 #if _MSVCP_VER >= 80
-DEFINE_RTTI_DATA(logic_error, 0, ".?AVlogic_error@std@@", exception_rtti_base_descriptor)
+DEFINE_RTTI_DATA1(logic_error, 0, &exception_rtti_base_descriptor, ".?AVlogic_error@std@@")
 #else
-DEFINE_RTTI_DATA(logic_error, 0, ".?AVlogic_error@@", exception_rtti_base_descriptor)
-#endif
-DEFINE_CXX_TYPE(logic_error, MSVCP_logic_error_dtor, exception_cxx_type_info)
+DEFINE_RTTI_DATA1(logic_error, 0, &exception_rtti_base_descriptor, ".?AVlogic_error@@")
+#endif
+DEFINE_CXX_TYPE_INFO(logic_error)
 
 /* length_error class data */
 typedef logic_error length_error;
@@ -457,8 +457,8 @@
     return length_error_copy_ctor(this, assign);
 }
 
-DEFINE_RTTI_DATA(length_error, 0, ".?AVlength_error@std@@", logic_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(length_error, MSVCP_logic_error_dtor, logic_error_cxx_type_info, exception_cxx_type_info)
+DEFINE_RTTI_DATA2(length_error, 0, &logic_error_rtti_base_descriptor, &exception_rtti_base_descriptor, ".?AVlength_error@std@@")
+DEFINE_CXX_DATA2(length_error, &logic_error_cxx_type_info, &exception_cxx_type_info, MSVCP_logic_error_dtor)
 
 /* out_of_range class data */
 typedef logic_error out_of_range;
@@ -504,8 +504,8 @@
     return out_of_range_copy_ctor(this, assign);
 }
 
-DEFINE_RTTI_DATA(out_of_range, 0, ".?AVout_of_range@std@@", logic_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(out_of_range, MSVCP_logic_error_dtor, logic_error_cxx_type_info, exception_cxx_type_info)
+DEFINE_RTTI_DATA2(out_of_range, 0, &logic_error_rtti_base_descriptor, &exception_rtti_base_descriptor, ".?AVout_of_range@std@@")
+DEFINE_CXX_DATA2(out_of_range, &logic_error_cxx_type_info, &exception_cxx_type_info, MSVCP_logic_error_dtor)
 
 /* invalid_argument class data */
 typedef logic_error invalid_argument;
@@ -528,8 +528,8 @@
     return this;
 }
 
-DEFINE_RTTI_DATA(invalid_argument, 0, ".?AVinvalid_argument@std@@", logic_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(invalid_argument, MSVCP_logic_error_dtor, logic_error_cxx_type_info,  exception_cxx_type_info)
+DEFINE_RTTI_DATA2(invalid_argument, 0, &logic_error_rtti_base_descriptor, &exception_rtti_base_descriptor, ".?AVinvalid_argument@std@@")
+DEFINE_CXX_DATA2(invalid_argument, &logic_error_cxx_type_info,  &exception_cxx_type_info, MSVCP_logic_error_dtor)
 
 /* runtime_error class data */
 typedef struct {
@@ -640,8 +640,8 @@
 #endif
 }
 
-DEFINE_RTTI_DATA(runtime_error, 0, ".?AVruntime_error@std@@", exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(runtime_error, MSVCP_runtime_error_dtor, exception_cxx_type_info)
+DEFINE_RTTI_DATA1(runtime_error, 0, &exception_rtti_base_descriptor, ".?AVruntime_error@std@@")
+DEFINE_CXX_DATA1(runtime_error, &exception_cxx_type_info, MSVCP_runtime_error_dtor)
 
 /* failure class data */
 typedef struct {
@@ -723,31 +723,40 @@
 #endif
 
 #if _MSVCP_VER > 110
-DEFINE_RTTI_DATA(_System_error, 0, ".?AV_System_error@std@@",
-        runtime_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_RTTI_DATA(system_error, 0, ".?AVsystem_error@std@@", _System_error_rtti_base_descriptor,
-        runtime_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_RTTI_DATA(failure, 0, ".?AVfailure@ios_base@std@@", system_error_rtti_base_descriptor,
-        _System_error_rtti_base_descriptor, runtime_error_rtti_base_descriptor,
-        exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(_System_error, MSVCP_runtime_error_dtor, runtime_error_cxx_type_info, exception_cxx_type_info)
-DEFINE_CXX_TYPE(system_error, MSVCP_runtime_error_dtor, _System_error_cxx_type_info,
-        runtime_error_cxx_type_info, exception_cxx_type_info)
-DEFINE_CXX_TYPE(failure, MSVCP_runtime_error_dtor, system_error_cxx_type_info,
-        _System_error_cxx_type_info, runtime_error_cxx_type_info,
-        exception_cxx_type_info)
+DEFINE_RTTI_DATA2(_System_error, 0, &runtime_error_rtti_base_descriptor,
+        &exception_rtti_base_descriptor, ".?AV_System_error@std@@")
+DEFINE_RTTI_DATA3(system_error, 0, &_System_error_rtti_base_descriptor,
+        &runtime_error_rtti_base_descriptor, &exception_rtti_base_descriptor,
+        ".?AVsystem_error@std@@")
+DEFINE_RTTI_DATA4(failure, 0, &system_error_rtti_base_descriptor,
+        &_System_error_rtti_base_descriptor, &runtime_error_rtti_base_descriptor,
+        &exception_rtti_base_descriptor, ".?AVfailure@ios_base@std@@")
+DEFINE_CXX_TYPE_INFO(_System_error)
+DEFINE_CXX_DATA3(system_error, &_System_error_cxx_type_info,
+        &runtime_error_cxx_type_info, &exception_cxx_type_info,
+        MSVCP_runtime_error_dtor)
+DEFINE_CXX_DATA4(failure, &system_error_cxx_type_info,
+        &_System_error_cxx_type_info, &runtime_error_cxx_type_info,
+        &exception_cxx_type_info, MSVCP_runtime_error_dtor)
 #elif _MSVCP_VER > 90
-DEFINE_RTTI_DATA(system_error, 0, ".?AVsystem_error@std@@",
-        runtime_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_RTTI_DATA(failure, 0, ".?AVfailure@ios_base@std@@",
-        system_error_rtti_base_descriptor, runtime_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(system_error, MSVCP_runtime_error_dtor, runtime_error_cxx_type_info, exception_cxx_type_info)
-DEFINE_CXX_TYPE(failure, MSVCP_runtime_error_dtor, system_error_cxx_type_info,
-        runtime_error_cxx_type_info, exception_cxx_type_info)
+DEFINE_RTTI_DATA2(system_error, 0, &runtime_error_rtti_base_descriptor,
+        &exception_rtti_base_descriptor, ".?AVsystem_error@std@@")
+DEFINE_RTTI_DATA3(failure, 0, &system_error_rtti_base_descriptor,
+        &runtime_error_rtti_base_descriptor, &exception_rtti_base_descriptor,
+        ".?AVfailure@ios_base@std@@")
+#if _MSVCP_VER == 100
+DEFINE_CXX_TYPE_INFO(system_error);
 #else
-DEFINE_RTTI_DATA(failure, 0, ".?AVfailure@ios_base@std@@",
-        runtime_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(failure, MSVCP_runtime_error_dtor, runtime_error_cxx_type_info, exception_cxx_type_info)
+DEFINE_CXX_DATA2(system_error, &runtime_error_cxx_type_info,
+        &exception_cxx_type_info, MSVCP_runtime_error_dtor)
+#endif
+DEFINE_CXX_DATA3(failure, &system_error_cxx_type_info, &runtime_error_cxx_type_info,
+        &exception_cxx_type_info, MSVCP_runtime_error_dtor)
+#else
+DEFINE_RTTI_DATA2(failure, 0, &runtime_error_rtti_base_descriptor,
+        &exception_rtti_base_descriptor, ".?AVfailure@ios_base@std@@")
+DEFINE_CXX_DATA2(failure, &runtime_error_cxx_type_info,
+        &exception_cxx_type_info, MSVCP_runtime_error_dtor)
 #endif
 
 /* bad_cast class data */
@@ -826,7 +835,7 @@
     return this;
 }
 
-DEFINE_RTTI_DATA(bad_cast, 0, ".?AVbad_cast@std@@", exception_rtti_base_descriptor)
+DEFINE_RTTI_DATA1(bad_cast, 0, &exception_rtti_base_descriptor, ".?AVbad_cast@std@@")
 
 /* range_error class data */
 typedef runtime_error range_error;
@@ -872,8 +881,8 @@
     return range_error_copy_ctor(this, assign);
 }
 
-DEFINE_RTTI_DATA(range_error, 0, ".?AVrange_error@std@@", runtime_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(range_error, MSVCP_runtime_error_dtor, runtime_error_cxx_type_info, exception_cxx_type_info)
+DEFINE_RTTI_DATA2(range_error, 0, &runtime_error_rtti_base_descriptor, &exception_rtti_base_descriptor, ".?AVrange_error@std@@")
+DEFINE_CXX_DATA2(range_error, &runtime_error_cxx_type_info, &exception_cxx_type_info, MSVCP_runtime_error_dtor)
 
 #if _MSVCP_VER > 90
 /* bad_function_call class data */
@@ -898,8 +907,8 @@
     return this;
 }
 
-DEFINE_RTTI_DATA(bad_function_call, 0, ".?AVbad_function_call@std@@", exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(bad_function_call, MSVCP_exception_dtor, exception_cxx_type_info)
+DEFINE_RTTI_DATA1(bad_function_call, 0, &exception_rtti_base_descriptor, ".?AVbad_function_call@std@@")
+DEFINE_CXX_DATA1(bad_function_call, &exception_cxx_type_info, MSVCP_exception_dtor)
 #endif
 
 /* ?_Nomemory@std@@YAXXZ */
@@ -1051,10 +1060,11 @@
     return code >= 0 && code < ARRAY_SIZE(names) ? names[code] : NULL;
 }
 
-DEFINE_RTTI_DATA(future_error, 0, ".?AVfuture_error@std@@",
-        future_error_rtti_base_descriptor, logic_error_rtti_base_descriptor, exception_rtti_base_descriptor)
-DEFINE_CXX_TYPE(future_error, MSVCP_logic_error_dtor, logic_error_cxx_type_info,
-        logic_error_cxx_type_info, exception_cxx_type_info)
+DEFINE_RTTI_DATA3(future_error, 0, &future_error_rtti_base_descriptor,
+        &logic_error_rtti_base_descriptor, &exception_rtti_base_descriptor,
+        ".?AVfuture_error@std@@")
+DEFINE_CXX_DATA3(future_error, &logic_error_cxx_type_info, &logic_error_cxx_type_info,
+        &exception_cxx_type_info, MSVCP_logic_error_dtor)
 
 /* ?_Throw_future_error@std@@YAXABVerror_code@1@@Z */
 /* ?_Throw_future_error@std@@YAXAEBVerror_code@1@@Z */
@@ -1179,6 +1189,11 @@
                    __ASM_CFI(".cfi_def_cfa %esp,4\n\t")
                    __ASM_CFI(".cfi_same_value %ebp\n\t")
                    "ret" )
+extern void call_dtor( void *func, void *this );
+__ASM_GLOBAL_FUNC( call_dtor,
+                   "movl 8(%esp),%ecx\n\t"
+                   "call *4(%esp)\n\t"
+                   "ret" )
 #else
 static inline void call_copy_ctor( void *func, void *this, void *src, int has_vbase )
 {
@@ -1188,36 +1203,11 @@
     else
         ((void (__thiscall*)(void*, void*))func)(this, src);
 }
-#endif
-
-/* copy the exception object where the catch block wants it */
-static inline void copy_exception( void *object, void **dest, UINT catch_flags,
-                                   const cxx_type_info *type, uintptr_t base )
-{
-    if (type->flags & CLASS_IS_SIMPLE_TYPE)
-    {
-        if (type->flags & CLASS_IS_WINRT && *(IUnknown**)object)
-            IUnknown_AddRef(*(IUnknown**)object);
-        memmove( dest, object, type->size );
-        /* if it is a pointer, adjust it */
-        if (type->size == sizeof(void*)) *dest = get_this_pointer( &type->offsets, *dest );
-    }
-    else  /* copy the object */
-    {
-        if (type->copy_ctor)
-        {
-            call_copy_ctor( cxx_rva( type->copy_ctor, base ), dest,
-                            get_this_pointer( &type->offsets, object ),
-                            (type->flags & CLASS_HAS_VIRTUAL_BASE_CLASS) );
-        }
-        else
-        {
-            if (type->flags & CLASS_IS_WINRT && *(IUnknown**)object)
-                IUnknown_AddRef(*(IUnknown**)object);
-            memmove( dest, get_this_pointer( &type->offsets, object ), type->size );
-        }
-    }
-}
+static inline void call_dtor( void *func, void *this )
+{
+    ((void (__thiscall*)(void*))func)( this );
+}
+#endif
 
 int __cdecl __uncaught_exceptions(void)
 {
@@ -1251,9 +1241,11 @@
     {
         if (ep->rec->ExceptionCode == CXX_EXCEPTION)
         {
+            const cxx_exception_type *type = (void*)ep->rec->ExceptionInformation[2];
             void *obj = (void*)ep->rec->ExceptionInformation[1];
-
-            __DestructExceptionObject(ep->rec);
+            uintptr_t base = rtti_rva_base( type );
+
+            if (type && type->destructor) call_dtor( rtti_rva(type->destructor, base), obj );
             HeapFree(GetProcessHeap(), 0, obj);
         }
 
@@ -1330,12 +1322,23 @@
     {
         void *obj = (void*)ep->rec->ExceptionInformation[1];
         const cxx_exception_type *et = (void*)ep->rec->ExceptionInformation[2];
-        uintptr_t base = cxx_rva_base( et );
-        const cxx_type_info_table *table = cxx_rva( et->type_info_table, base );
-        const cxx_type_info *ti = cxx_rva( table->info[0], base );
+        uintptr_t base = rtti_rva_base( et );
+        const cxx_type_info_table *table = rtti_rva( et->type_info_table, base );
+        const cxx_type_info *ti = rtti_rva( table->info[0], base );
         void **data = HeapAlloc(GetProcessHeap(), 0, ti->size);
 
-        copy_exception(obj, data, 0, ti, base);
+        if (ti->flags & CLASS_IS_SIMPLE_TYPE)
+        {
+            memcpy(data, obj, ti->size);
+            if (ti->size == sizeof(void *)) *data = get_this_pointer(&ti->offsets, *data);
+        }
+        else if (ti->copy_ctor)
+        {
+            call_copy_ctor(rtti_rva(ti->copy_ctor, base), data, get_this_pointer(&ti->offsets, obj),
+                    ti->flags & CLASS_HAS_VIRTUAL_BASE_CLASS);
+        }
+        else
+            memcpy(data, get_this_pointer(&ti->offsets, obj), ti->size);
         ep->rec->ExceptionInformation[1] = (ULONG_PTR)data;
     }
     return;
@@ -1360,7 +1363,7 @@
     const cxx_type_info_table *table;
     const cxx_type_info *ti;
     void **data;
-    uintptr_t base = cxx_rva_base( type );
+    uintptr_t base = rtti_rva_base( type );
 
     __ExceptionPtrDestroy(ep);
 
@@ -1376,10 +1379,21 @@
     ep->rec->ExceptionInformation[2] = (ULONG_PTR)type;
     if (CXX_EXCEPTION_PARAMS == 4) ep->rec->ExceptionInformation[3] = base;
 
-    table = cxx_rva( type->type_info_table, base );
-    ti = cxx_rva( table->info[0], base );
+    table = rtti_rva( type->type_info_table, base );
+    ti = rtti_rva( table->info[0], base );
     data = HeapAlloc(GetProcessHeap(), 0, ti->size);
-    copy_exception(object, data, 0, ti, base);
+    if (ti->flags & CLASS_IS_SIMPLE_TYPE)
+    {
+        memcpy(data, object, ti->size);
+        if (ti->size == sizeof(void *)) *data = get_this_pointer(&ti->offsets, *data);
+    }
+    else if (ti->copy_ctor)
+    {
+        call_copy_ctor( rtti_rva(ti->copy_ctor, base), data, get_this_pointer(&ti->offsets, object),
+                ti->flags & CLASS_HAS_VIRTUAL_BASE_CLASS);
+    }
+    else
+        memcpy(data, get_this_pointer(&ti->offsets, object), ti->size);
     ep->rec->ExceptionInformation[1] = (ULONG_PTR)data;
 }
 
@@ -1483,45 +1497,51 @@
 
 void init_exception(void *base)
 {
-    INIT_RTTI(type_info, base);
-    INIT_RTTI(exception, base);
-    INIT_RTTI(bad_alloc, base);
-    INIT_RTTI(logic_error, base);
-    INIT_RTTI(length_error, base);
-    INIT_RTTI(out_of_range, base);
-    INIT_RTTI(invalid_argument, base);
-    INIT_RTTI(runtime_error, base);
+#ifdef RTTI_USE_RVA
+    init_type_info_rtti(base);
+    init_exception_rtti(base);
+    init_bad_alloc_rtti(base);
+    init_logic_error_rtti(base);
+    init_length_error_rtti(base);
+    init_out_of_range_rtti(base);
+    init_invalid_argument_rtti(base);
+    init_runtime_error_rtti(base);
 #if _MSVCP_VER >= 110
-    INIT_RTTI(future_error, base);
+    init_future_error_rtti(base);
 #endif
 #if _MSVCP_VER > 110
-    INIT_RTTI(_System_error, base);
+    init__System_error_rtti(base);
 #endif
 #if _MSVCP_VER > 90
-    INIT_RTTI(system_error, base);
-    INIT_RTTI(bad_function_call, base);
-#endif
-    INIT_RTTI(failure, base);
-    INIT_RTTI(bad_cast, base);
-    INIT_RTTI(range_error, base);
-
-    INIT_CXX_TYPE(exception, base);
-    INIT_CXX_TYPE(bad_alloc, base);
-    INIT_CXX_TYPE(logic_error, base);
-    INIT_CXX_TYPE(length_error, base);
-    INIT_CXX_TYPE(out_of_range, base);
-    INIT_CXX_TYPE(invalid_argument, base);
-    INIT_CXX_TYPE(runtime_error, base);
+    init_system_error_rtti(base);
+    init_bad_function_call_rtti(base);
+#endif
+    init_failure_rtti(base);
+    init_bad_cast_rtti(base);
+    init_range_error_rtti(base);
+
+    init_exception_cxx(base);
+    init_bad_alloc_cxx(base);
+    init_logic_error_cxx_type_info(base);
+    init_length_error_cxx(base);
+    init_out_of_range_cxx(base);
+    init_invalid_argument_cxx(base);
+    init_runtime_error_cxx(base);
 #if _MSVCP_VER >= 110
-    INIT_CXX_TYPE(future_error, base);
+    init_future_error_cxx(base);
 #endif
 #if _MSVCP_VER > 110
-    INIT_CXX_TYPE(_System_error, base);
+    init__System_error_cxx_type_info(base);
+#endif
+#if _MSVCP_VER == 100
+    init_system_error_cxx_type_info(base);
+#elif _MSVCP_VER > 100
+    init_system_error_cxx(base);
 #endif
 #if _MSVCP_VER > 90
-    INIT_CXX_TYPE(system_error, base);
-    INIT_CXX_TYPE(bad_function_call, base);
-#endif
-    INIT_CXX_TYPE(failure, base);
-    INIT_CXX_TYPE(range_error, base);
-}
+    init_bad_function_call_cxx(base);
+#endif
+    init_failure_cxx(base);
+    init_range_error_cxx(base);
+#endif
+}
