diff --git a/dlls/wined3d/cs.c b/dlls/wined3d/cs.c
--- a/dlls/wined3d/cs.c
+++ b/dlls/wined3d/cs.c
@@ -118,11 +118,11 @@
     WINED3D_CS_OP_SET_DEPTH_BOUNDS,
     WINED3D_CS_OP_SET_RENDER_STATE,
     WINED3D_CS_OP_SET_TEXTURE_STATE,
+    WINED3D_CS_OP_SET_TRANSFORM,
+    WINED3D_CS_OP_SET_CLIP_PLANE,
     WINED3D_CS_OP_SET_COLOR_KEY,
     WINED3D_CS_OP_SET_LIGHT,
     WINED3D_CS_OP_SET_LIGHT_ENABLE,
-    WINED3D_CS_OP_SET_EXTRA_VS_ARGS,
-    WINED3D_CS_OP_SET_EXTRA_PS_ARGS,
     WINED3D_CS_OP_SET_FEATURE_LEVEL,
     WINED3D_CS_OP_PUSH_CONSTANTS,
     WINED3D_CS_OP_RESET_STATE,
@@ -140,7 +140,6 @@
     WINED3D_CS_OP_CLEAR_UNORDERED_ACCESS_VIEW,
     WINED3D_CS_OP_COPY_UAV_COUNTER,
     WINED3D_CS_OP_GENERATE_MIPMAPS,
-    WINED3D_CS_OP_DECODE,
     WINED3D_CS_OP_EXECUTE_COMMAND_LIST,
     WINED3D_CS_OP_STOP,
 };
@@ -379,6 +378,20 @@
     DWORD value;
 };
 
+struct wined3d_cs_set_transform
+{
+    enum wined3d_cs_op opcode;
+    enum wined3d_transform_state state;
+    struct wined3d_matrix matrix;
+};
+
+struct wined3d_cs_set_clip_plane
+{
+    enum wined3d_cs_op opcode;
+    UINT plane_idx;
+    struct wined3d_vec4 plane;
+};
+
 struct wined3d_cs_set_light
 {
     enum wined3d_cs_op opcode;
@@ -390,18 +403,6 @@
     enum wined3d_cs_op opcode;
     unsigned int idx;
     BOOL enable;
-};
-
-struct wined3d_cs_set_extra_vs_args
-{
-    enum wined3d_cs_op opcode;
-    struct wined3d_extra_vs_args args;
-};
-
-struct wined3d_cs_set_extra_ps_args
-{
-    enum wined3d_cs_op opcode;
-    struct wined3d_extra_ps_args args;
 };
 
 struct wined3d_cs_set_feature_level
@@ -529,14 +530,6 @@
 {
     enum wined3d_cs_op opcode;
     struct wined3d_shader_resource_view *view;
-};
-
-struct wined3d_cs_decode
-{
-    enum wined3d_cs_op opcode;
-    struct wined3d_decoder *decoder;
-    struct wined3d_decoder_output_view *output_view;
-    unsigned int bitstream_size, slice_control_size;
 };
 
 struct wined3d_cs_execute_command_list
@@ -611,11 +604,11 @@
         WINED3D_TO_STR(WINED3D_CS_OP_SET_DEPTH_BOUNDS);
         WINED3D_TO_STR(WINED3D_CS_OP_SET_RENDER_STATE);
         WINED3D_TO_STR(WINED3D_CS_OP_SET_TEXTURE_STATE);
+        WINED3D_TO_STR(WINED3D_CS_OP_SET_TRANSFORM);
+        WINED3D_TO_STR(WINED3D_CS_OP_SET_CLIP_PLANE);
         WINED3D_TO_STR(WINED3D_CS_OP_SET_COLOR_KEY);
         WINED3D_TO_STR(WINED3D_CS_OP_SET_LIGHT);
         WINED3D_TO_STR(WINED3D_CS_OP_SET_LIGHT_ENABLE);
-        WINED3D_TO_STR(WINED3D_CS_OP_SET_EXTRA_PS_ARGS);
-        WINED3D_TO_STR(WINED3D_CS_OP_SET_EXTRA_VS_ARGS);
         WINED3D_TO_STR(WINED3D_CS_OP_SET_FEATURE_LEVEL);
         WINED3D_TO_STR(WINED3D_CS_OP_PUSH_CONSTANTS);
         WINED3D_TO_STR(WINED3D_CS_OP_RESET_STATE);
@@ -633,7 +626,6 @@
         WINED3D_TO_STR(WINED3D_CS_OP_CLEAR_UNORDERED_ACCESS_VIEW);
         WINED3D_TO_STR(WINED3D_CS_OP_COPY_UAV_COUNTER);
         WINED3D_TO_STR(WINED3D_CS_OP_GENERATE_MIPMAPS);
-        WINED3D_TO_STR(WINED3D_CS_OP_DECODE);
         WINED3D_TO_STR(WINED3D_CS_OP_EXECUTE_COMMAND_LIST);
         WINED3D_TO_STR(WINED3D_CS_OP_STOP);
 #undef WINED3D_TO_STR
@@ -1137,7 +1129,7 @@
 }
 
 static void reference_graphics_pipeline_resources(struct wined3d_device_context *context,
-        bool indexed, const struct wined3d_d3d_info *d3d_info)
+        BOOL indexed, const struct wined3d_d3d_info *d3d_info)
 {
     const struct wined3d_state *state = context->state;
     unsigned int i;
@@ -1204,13 +1196,6 @@
     struct wined3d_cs_draw *op;
 
     wined3d_device_context_lock(context);
-
-    if (indexed && !state->index_buffer)
-    {
-        wined3d_device_context_unlock(context);
-        return;
-    }
-
     op = wined3d_device_context_require_space(context, sizeof(*op), WINED3D_CS_QUEUE_DEFAULT);
     op->opcode = WINED3D_CS_OP_DRAW;
     op->primitive_type = state->primitive_type;
@@ -1343,12 +1328,12 @@
             device_invalidate_state(device, STATE_SHADER(WINED3D_SHADER_TYPE_PIXEL));
 
         if (!(device->adapter->d3d_info.wined3d_creation_flags & WINED3D_SRGB_READ_WRITE_CONTROL)
-                || cs->state.extra_ps_args.srgb_write)
+                || cs->state.render_states[WINED3D_RS_SRGBWRITEENABLE])
         {
             prev_srgb_write = prev && prev->format_caps & WINED3D_FORMAT_CAP_SRGB_WRITE;
             curr_srgb_write = view && view->format_caps & WINED3D_FORMAT_CAP_SRGB_WRITE;
             if (prev_srgb_write != curr_srgb_write)
-                device_invalidate_state(device, STATE_SHADER(WINED3D_SHADER_TYPE_PIXEL));
+                device_invalidate_state(device, STATE_RENDER(WINED3D_RS_SRGBWRITEENABLE));
         }
     }
 
@@ -1578,7 +1563,7 @@
 static bool texture_binding_might_invalidate_ps(struct wined3d_shader_resource_view *view,
         struct wined3d_shader_resource_view *prev, const struct wined3d_d3d_info *d3d_info)
 {
-    unsigned int old_usage, new_usage;
+    unsigned int old_usage, new_usage, old_caps, new_caps;
     const struct wined3d_format *old_format, *new_format;
 
     if (!prev)
@@ -1593,7 +1578,9 @@
 
     old_format = prev->resource->format;
     new_format = view->resource->format;
-    if ((old_format->attrs & WINED3D_FORMAT_ATTR_SHADOW) != (new_format->attrs & WINED3D_FORMAT_ATTR_SHADOW))
+    old_caps = prev->resource->format_caps;
+    new_caps = view->resource->format_caps;
+    if ((old_caps & WINED3D_FORMAT_CAP_SHADOW) != (new_caps & WINED3D_FORMAT_CAP_SHADOW))
         return true;
 
     if (is_same_fixup(old_format->color_fixup, new_format->color_fixup))
@@ -1938,6 +1925,52 @@
     wined3d_device_context_submit(context, WINED3D_CS_QUEUE_DEFAULT);
 }
 
+static void wined3d_cs_exec_set_transform(struct wined3d_cs *cs, const void *data)
+{
+    const struct wined3d_cs_set_transform *op = data;
+
+    cs->state.transforms[op->state] = op->matrix;
+    /* Fog behaviour depends on the projection matrix. */
+    if (op->state == WINED3D_TS_PROJECTION
+            && cs->state.render_states[WINED3D_RS_FOGENABLE]
+            && cs->state.render_states[WINED3D_RS_FOGTABLEMODE] != WINED3D_FOG_NONE)
+        device_invalidate_state(cs->c.device, STATE_SHADER(WINED3D_SHADER_TYPE_VERTEX));
+}
+
+void wined3d_device_context_emit_set_transform(struct wined3d_device_context *context,
+        enum wined3d_transform_state state, const struct wined3d_matrix *matrix)
+{
+    struct wined3d_cs_set_transform *op;
+
+    op = wined3d_device_context_require_space(context, sizeof(*op), WINED3D_CS_QUEUE_DEFAULT);
+    op->opcode = WINED3D_CS_OP_SET_TRANSFORM;
+    op->state = state;
+    op->matrix = *matrix;
+
+    wined3d_device_context_submit(context, WINED3D_CS_QUEUE_DEFAULT);
+}
+
+static void wined3d_cs_exec_set_clip_plane(struct wined3d_cs *cs, const void *data)
+{
+    const struct wined3d_cs_set_clip_plane *op = data;
+
+    cs->state.clip_planes[op->plane_idx] = op->plane;
+    device_invalidate_state(cs->c.device, STATE_CLIPPLANE(op->plane_idx));
+}
+
+void wined3d_device_context_emit_set_clip_plane(struct wined3d_device_context *context,
+        unsigned int plane_idx, const struct wined3d_vec4 *plane)
+{
+    struct wined3d_cs_set_clip_plane *op;
+
+    op = wined3d_device_context_require_space(context, sizeof(*op), WINED3D_CS_QUEUE_DEFAULT);
+    op->opcode = WINED3D_CS_OP_SET_CLIP_PLANE;
+    op->plane_idx = plane_idx;
+    op->plane = *plane;
+
+    wined3d_device_context_submit(context, WINED3D_CS_QUEUE_DEFAULT);
+}
+
 static void wined3d_cs_exec_set_color_key(struct wined3d_cs *cs, const void *data)
 {
     const struct wined3d_cs_set_color_key *op = data;
@@ -2086,46 +2119,6 @@
     wined3d_device_context_submit(context, WINED3D_CS_QUEUE_DEFAULT);
 }
 
-static void wined3d_cs_exec_set_extra_vs_args(struct wined3d_cs *cs, const void *data)
-{
-    const struct wined3d_cs_set_extra_vs_args *op = data;
-
-    cs->state.extra_vs_args = op->args;
-    device_invalidate_state(cs->c.device, STATE_SHADER(WINED3D_SHADER_TYPE_VERTEX));
-}
-
-void wined3d_device_context_emit_set_extra_vs_args(struct wined3d_device_context *context,
-        const struct wined3d_extra_vs_args *args)
-{
-    struct wined3d_cs_set_extra_vs_args *op;
-
-    op = wined3d_device_context_require_space(context, sizeof(*op), WINED3D_CS_QUEUE_DEFAULT);
-    op->opcode = WINED3D_CS_OP_SET_EXTRA_VS_ARGS;
-    op->args = *args;
-
-    wined3d_device_context_submit(context, WINED3D_CS_QUEUE_DEFAULT);
-}
-
-static void wined3d_cs_exec_set_extra_ps_args(struct wined3d_cs *cs, const void *data)
-{
-    const struct wined3d_cs_set_extra_ps_args *op = data;
-
-    cs->state.extra_ps_args = op->args;
-    device_invalidate_state(cs->c.device, STATE_SHADER(WINED3D_SHADER_TYPE_PIXEL));
-}
-
-void wined3d_device_context_emit_set_extra_ps_args(struct wined3d_device_context *context,
-        const struct wined3d_extra_ps_args *args)
-{
-    struct wined3d_cs_set_extra_ps_args *op;
-
-    op = wined3d_device_context_require_space(context, sizeof(*op), WINED3D_CS_QUEUE_DEFAULT);
-    op->opcode = WINED3D_CS_OP_SET_EXTRA_PS_ARGS;
-    op->args = *args;
-
-    wined3d_device_context_submit(context, WINED3D_CS_QUEUE_DEFAULT);
-}
-
 static void wined3d_cs_exec_set_feature_level(struct wined3d_cs *cs, const void *data)
 {
     const struct wined3d_cs_set_feature_level *op = data;
@@ -2150,17 +2143,18 @@
     size_t size;
     unsigned int max_count;
     enum wined3d_shader_type shader_type;
+    enum vkd3d_shader_d3dbc_constant_register shader_binding;
 }
 wined3d_cs_push_constant_info[] =
 {
-    [WINED3D_PUSH_CONSTANTS_VS_F]       = {sizeof(struct wined3d_vec4),  WINED3D_MAX_VS_CONSTS_F, WINED3D_SHADER_TYPE_VERTEX},
-    [WINED3D_PUSH_CONSTANTS_PS_F]       = {sizeof(struct wined3d_vec4),  WINED3D_MAX_PS_CONSTS_F, WINED3D_SHADER_TYPE_PIXEL},
-    [WINED3D_PUSH_CONSTANTS_VS_I]       = {sizeof(struct wined3d_ivec4), WINED3D_MAX_CONSTS_I,    WINED3D_SHADER_TYPE_VERTEX},
-    [WINED3D_PUSH_CONSTANTS_PS_I]       = {sizeof(struct wined3d_ivec4), WINED3D_MAX_CONSTS_I,    WINED3D_SHADER_TYPE_PIXEL},
-    [WINED3D_PUSH_CONSTANTS_VS_B]       = {sizeof(BOOL),                 WINED3D_MAX_CONSTS_B,    WINED3D_SHADER_TYPE_VERTEX},
-    [WINED3D_PUSH_CONSTANTS_PS_B]       = {sizeof(BOOL),                 WINED3D_MAX_CONSTS_B,    WINED3D_SHADER_TYPE_PIXEL},
-    [WINED3D_PUSH_CONSTANTS_VS_FFP]     = {1, sizeof(struct wined3d_ffp_vs_constants),            WINED3D_SHADER_TYPE_VERTEX},
-    [WINED3D_PUSH_CONSTANTS_PS_FFP]     = {1, sizeof(struct wined3d_ffp_ps_constants),            WINED3D_SHADER_TYPE_PIXEL},
+    [WINED3D_PUSH_CONSTANTS_VS_F]       = {sizeof(struct wined3d_vec4),  WINED3D_MAX_VS_CONSTS_F, WINED3D_SHADER_TYPE_VERTEX, VKD3D_SHADER_D3DBC_FLOAT_CONSTANT_REGISTER},
+    [WINED3D_PUSH_CONSTANTS_PS_F]       = {sizeof(struct wined3d_vec4),  WINED3D_MAX_PS_CONSTS_F, WINED3D_SHADER_TYPE_PIXEL,  VKD3D_SHADER_D3DBC_FLOAT_CONSTANT_REGISTER},
+    [WINED3D_PUSH_CONSTANTS_VS_I]       = {sizeof(struct wined3d_ivec4), WINED3D_MAX_CONSTS_I,    WINED3D_SHADER_TYPE_VERTEX, VKD3D_SHADER_D3DBC_INT_CONSTANT_REGISTER},
+    [WINED3D_PUSH_CONSTANTS_PS_I]       = {sizeof(struct wined3d_ivec4), WINED3D_MAX_CONSTS_I,    WINED3D_SHADER_TYPE_PIXEL,  VKD3D_SHADER_D3DBC_INT_CONSTANT_REGISTER},
+    [WINED3D_PUSH_CONSTANTS_VS_B]       = {sizeof(BOOL),                 WINED3D_MAX_CONSTS_B,    WINED3D_SHADER_TYPE_VERTEX, VKD3D_SHADER_D3DBC_BOOL_CONSTANT_REGISTER},
+    [WINED3D_PUSH_CONSTANTS_PS_B]       = {sizeof(BOOL),                 WINED3D_MAX_CONSTS_B,    WINED3D_SHADER_TYPE_PIXEL,  VKD3D_SHADER_D3DBC_BOOL_CONSTANT_REGISTER},
+    [WINED3D_PUSH_CONSTANTS_VS_FFP]     = {1, sizeof(struct wined3d_ffp_vs_constants),            WINED3D_SHADER_TYPE_VERTEX, VKD3D_SHADER_D3DBC_FLOAT_CONSTANT_REGISTER},
+    [WINED3D_PUSH_CONSTANTS_PS_FFP]     = {1, sizeof(struct wined3d_ffp_ps_constants),            WINED3D_SHADER_TYPE_PIXEL,  VKD3D_SHADER_D3DBC_FLOAT_CONSTANT_REGISTER},
 };
 
 static bool prepare_push_constant_buffer(struct wined3d_device_context *context, enum wined3d_push_constants type)
@@ -2188,6 +2182,13 @@
     {
         ERR("Failed to create push constant buffer, hr %#lx.\n", hr);
         return false;
+    }
+
+    if (gpu)
+    {
+        struct wined3d_constant_buffer_state state = {.buffer = device->push_constants[type], .size = desc.byte_width};
+
+        wined3d_device_context_emit_set_constant_buffers(context, info->shader_type, info->shader_binding, 1, &state);
     }
 
     return true;
@@ -2786,29 +2787,8 @@
 
     if (context->ops->map_upload_bo(context, resource, sub_resource_idx, &map_desc, box, WINED3D_MAP_WRITE))
     {
-        const struct wined3d_format *format = resource->format;
-
-        if (format->attrs & WINED3D_FORMAT_ATTR_PLANAR)
-        {
-            unsigned int uv_height = format->uv_height;
-            unsigned int uv_width = format->uv_width;
-            const struct wined3d_format *plane_format;
-
-            plane_format = wined3d_get_format(context->device->adapter, format->plane_formats[0], 0);
-            wined3d_format_copy_data(plane_format, data, row_pitch, slice_pitch, map_desc.data, map_desc.row_pitch,
-                    map_desc.slice_pitch, box->right - box->left, box->bottom - box->top, box->back - box->front);
-            plane_format = wined3d_get_format(context->device->adapter, format->plane_formats[1], 0);
-            wined3d_format_copy_data(plane_format, (const uint8_t *)data + (row_pitch * (box->bottom - box->top)),
-                    row_pitch * 2 / uv_width, slice_pitch * 2 / uv_width / uv_height,
-                    (uint8_t *)map_desc.data + map_desc.slice_pitch,
-                    map_desc.row_pitch * 2 / uv_width, map_desc.slice_pitch * 2 / uv_width / uv_height,
-                    (box->right - box->left) / uv_width, (box->bottom - box->top) / uv_height, box->back - box->front);
-        }
-        else
-        {
-            wined3d_format_copy_data(format, data, row_pitch, slice_pitch, map_desc.data, map_desc.row_pitch,
-                    map_desc.slice_pitch, box->right - box->left, box->bottom - box->top, box->back - box->front);
-        }
+        wined3d_format_copy_data(resource->format, data, row_pitch, slice_pitch, map_desc.data, map_desc.row_pitch,
+                map_desc.slice_pitch, box->right - box->left, box->bottom - box->top, box->back - box->front);
         context->ops->unmap_upload_bo(context, resource, sub_resource_idx, &dummy_box, &bo);
         wined3d_device_context_upload_bo(context, resource, sub_resource_idx,
                 box, &bo, map_desc.row_pitch, map_desc.slice_pitch);
@@ -2943,43 +2923,6 @@
     op->view = view;
 
     wined3d_device_context_reference_resource(context, view->resource);
-
-    wined3d_device_context_submit(context, WINED3D_CS_QUEUE_DEFAULT);
-}
-
-static void wined3d_cs_exec_decode(struct wined3d_cs *cs, const void *data)
-{
-    const struct wined3d_cs_decode *op = data;
-    struct wined3d_context *context;
-
-    context = context_acquire(cs->c.device, NULL, 0);
-    cs->c.device->adapter->decoder_ops->decode(context, op->decoder,
-            op->output_view, op->bitstream_size, op->slice_control_size);
-    context_release(context);
-}
-
-void wined3d_cs_emit_decode(struct wined3d_decoder *decoder, struct wined3d_decoder_output_view *output_view,
-        unsigned int bitstream_size, unsigned int slice_control_size)
-{
-    struct wined3d_device_context *context = &output_view->texture->resource.device->cs->c;
-    struct wined3d_cs_decode *op;
-
-    op = wined3d_device_context_require_space(context, sizeof(*op), WINED3D_CS_QUEUE_DEFAULT);
-    op->opcode = WINED3D_CS_OP_DECODE;
-    op->decoder = decoder;
-    op->output_view = output_view;
-    op->bitstream_size = bitstream_size;
-    op->slice_control_size = slice_control_size;
-
-    wined3d_device_context_reference_resource(context, &output_view->texture->resource);
-    wined3d_device_context_reference_resource(context,
-            wined3d_decoder_get_buffer(decoder, WINED3D_DECODER_BUFFER_BITSTREAM));
-    wined3d_device_context_reference_resource(context,
-            wined3d_decoder_get_buffer(decoder, WINED3D_DECODER_BUFFER_INVERSE_QUANTIZATION_MATRIX));
-    wined3d_device_context_reference_resource(context,
-            wined3d_decoder_get_buffer(decoder, WINED3D_DECODER_BUFFER_PICTURE_PARAMETERS));
-    wined3d_device_context_reference_resource(context,
-            wined3d_decoder_get_buffer(decoder, WINED3D_DECODER_BUFFER_SLICE_CONTROL));
 
     wined3d_device_context_submit(context, WINED3D_CS_QUEUE_DEFAULT);
 }
@@ -3031,11 +2974,11 @@
     /* WINED3D_CS_OP_SET_DEPTH_BOUNDS            */ wined3d_cs_exec_set_depth_bounds,
     /* WINED3D_CS_OP_SET_RENDER_STATE            */ wined3d_cs_exec_set_render_state,
     /* WINED3D_CS_OP_SET_TEXTURE_STATE           */ wined3d_cs_exec_set_texture_state,
+    /* WINED3D_CS_OP_SET_TRANSFORM               */ wined3d_cs_exec_set_transform,
+    /* WINED3D_CS_OP_SET_CLIP_PLANE              */ wined3d_cs_exec_set_clip_plane,
     /* WINED3D_CS_OP_SET_COLOR_KEY               */ wined3d_cs_exec_set_color_key,
     /* WINED3D_CS_OP_SET_LIGHT                   */ wined3d_cs_exec_set_light,
     /* WINED3D_CS_OP_SET_LIGHT_ENABLE            */ wined3d_cs_exec_set_light_enable,
-    /* WINED3D_CS_OP_SET_EXTRA_VS_ARGS           */ wined3d_cs_exec_set_extra_vs_args,
-    /* WINED3D_CS_OP_SET_EXTRA_PS_ARGS           */ wined3d_cs_exec_set_extra_ps_args,
     /* WINED3D_CS_OP_SET_FEATURE_LEVEL           */ wined3d_cs_exec_set_feature_level,
     /* WINED3D_CS_OP_PUSH_CONSTANTS              */ wined3d_cs_exec_push_constants,
     /* WINED3D_CS_OP_RESET_STATE                 */ wined3d_cs_exec_reset_state,
@@ -3053,7 +2996,6 @@
     /* WINED3D_CS_OP_CLEAR_UNORDERED_ACCESS_VIEW */ wined3d_cs_exec_clear_unordered_access_view,
     /* WINED3D_CS_OP_COPY_UAV_COUNTER            */ wined3d_cs_exec_copy_uav_counter,
     /* WINED3D_CS_OP_GENERATE_MIPMAPS            */ wined3d_cs_exec_generate_mipmaps,
-    /* WINED3D_CS_OP_DECODE                      */ wined3d_cs_exec_decode,
     /* WINED3D_CS_OP_EXECUTE_COMMAND_LIST        */ wined3d_cs_exec_execute_command_list,
 };
 
@@ -3129,27 +3071,6 @@
 
 static void wined3d_cs_st_finish(struct wined3d_device_context *context, enum wined3d_cs_queue_id queue_id)
 {
-}
-
-static void get_map_pitch(const struct wined3d_format *format, const struct wined3d_box *box,
-        struct wined3d_map_desc *map_desc, size_t *size)
-{
-    unsigned int height = box->bottom - box->top;
-    unsigned int width = box->right - box->left;
-    unsigned int depth = box->back - box->front;
-
-    wined3d_format_calculate_pitch(format, 1, width, height, &map_desc->row_pitch, &map_desc->slice_pitch);
-
-    if (format->attrs & WINED3D_FORMAT_ATTR_PLANAR)
-    {
-        *size = wined3d_format_calculate_size(format, 1, width, height, 1);
-    }
-    else
-    {
-        *size = (depth - 1) * map_desc->slice_pitch
-                + ((height - 1) / format->block_height) * map_desc->row_pitch
-                + ((width + format->block_width - 1) / format->block_width) * format->block_byte_count;
-    }
 }
 
 static bool wined3d_cs_map_upload_bo(struct wined3d_device_context *context, struct wined3d_resource *resource,
@@ -3242,7 +3163,12 @@
         return true;
     }
 
-    get_map_pitch(format, box, map_desc, &size);
+    wined3d_format_calculate_pitch(format, 1, box->right - box->left,
+            box->bottom - box->top, &map_desc->row_pitch, &map_desc->slice_pitch);
+
+    size = (box->back - box->front - 1) * map_desc->slice_pitch
+            + ((box->bottom - box->top - 1) / format->block_height) * map_desc->row_pitch
+            + ((box->right - box->left + format->block_width - 1) / format->block_width) * format->block_byte_count;
 
     if (!(map_desc->data = malloc(size)))
     {
@@ -4341,7 +4267,12 @@
     uint8_t *sysmem;
     size_t size;
 
-    get_map_pitch(format, box, map_desc, &size);
+    wined3d_format_calculate_pitch(format, 1, box->right - box->left,
+            box->bottom - box->top, &map_desc->row_pitch, &map_desc->slice_pitch);
+
+    size = (box->back - box->front - 1) * map_desc->slice_pitch
+            + ((box->bottom - box->top - 1) / format->block_height) * map_desc->row_pitch
+            + ((box->right - box->left + format->block_width - 1) / format->block_width) * format->block_byte_count;
 
     if (!(flags & WINED3D_MAP_WRITE))
     {
