diff --git a/dlls/d3d11/state.c b/dlls/d3d11/state.c
--- a/dlls/d3d11/state.c
+++ b/dlls/d3d11/state.c
@@ -64,7 +64,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(state->device);
+        ID3D11Device2_AddRef(state->device);
         wined3d_blend_state_incref(state->wined3d_state);
     }
 
@@ -80,9 +80,9 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = state->device;
+        ID3D11Device2 *device = state->device;
         wined3d_blend_state_decref(state->wined3d_state);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -223,7 +223,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(state->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(state->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_blend_state_GetPrivateData(ID3D10BlendState1 *iface,
@@ -321,7 +321,7 @@
 static void STDMETHODCALLTYPE d3d_blend_state_wined3d_object_destroyed(void *parent)
 {
     struct d3d_blend_state *state = parent;
-    struct d3d_device *device = impl_from_ID3D11Device5(state->device);
+    struct d3d_device *device = impl_from_ID3D11Device2(state->device);
 
     wine_rb_remove(&device->blend_states, &state->entry);
     wined3d_private_store_cleanup(&state->private_store);
@@ -458,7 +458,7 @@
     }
     wined3d_mutex_unlock();
 
-    ID3D11Device5_AddRef(object->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(object->device = &device->ID3D11Device2_iface);
 
     TRACE("Created blend state %p.\n", object);
     *state = object;
@@ -525,7 +525,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(state->device);
+        ID3D11Device2_AddRef(state->device);
         wined3d_depth_stencil_state_incref(state->wined3d_state);
     }
 
@@ -541,10 +541,10 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = state->device;
+        ID3D11Device2 *device = state->device;
 
         wined3d_depth_stencil_state_decref(state->wined3d_state);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -661,7 +661,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(state->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(state->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_depthstencil_state_GetPrivateData(ID3D10DepthStencilState *iface,
@@ -726,7 +726,7 @@
 static void STDMETHODCALLTYPE d3d_depthstencil_state_wined3d_object_destroyed(void *parent)
 {
     struct d3d_depthstencil_state *state = parent;
-    struct d3d_device *device = impl_from_ID3D11Device5(state->device);
+    struct d3d_device *device = impl_from_ID3D11Device2(state->device);
 
     wine_rb_remove(&device->depthstencil_states, &state->entry);
     wined3d_private_store_cleanup(&state->private_store);
@@ -859,7 +859,7 @@
     }
     wined3d_mutex_unlock();
 
-    ID3D11Device5_AddRef(object->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(object->device = &device->ID3D11Device2_iface);
 
     TRACE("Created depth/stencil state %p.\n", object);
     *state = object;
@@ -887,25 +887,24 @@
 
 /* ID3D11RasterizerState methods */
 
-static inline struct d3d_rasterizer_state *impl_from_ID3D11RasterizerState2(ID3D11RasterizerState2 *iface)
-{
-    return CONTAINING_RECORD(iface, struct d3d_rasterizer_state, ID3D11RasterizerState2_iface);
-}
-
-static HRESULT STDMETHODCALLTYPE d3d11_rasterizer_state_QueryInterface(ID3D11RasterizerState2 *iface,
+static inline struct d3d_rasterizer_state *impl_from_ID3D11RasterizerState1(ID3D11RasterizerState1 *iface)
+{
+    return CONTAINING_RECORD(iface, struct d3d_rasterizer_state, ID3D11RasterizerState1_iface);
+}
+
+static HRESULT STDMETHODCALLTYPE d3d11_rasterizer_state_QueryInterface(ID3D11RasterizerState1 *iface,
         REFIID riid, void **object)
 {
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
 
     TRACE("iface %p, riid %s, object %p.\n", iface, debugstr_guid(riid), object);
 
     if (IsEqualGUID(riid, &IID_ID3D11RasterizerState)
             || IsEqualGUID(riid, &IID_ID3D11RasterizerState1)
-            || IsEqualGUID(riid, &IID_ID3D11RasterizerState2)
             || IsEqualGUID(riid, &IID_ID3D11DeviceChild)
             || IsEqualGUID(riid, &IID_IUnknown))
     {
-        ID3D11RasterizerState2_AddRef(iface);
+        ID3D11RasterizerState1_AddRef(iface);
         *object = iface;
         return S_OK;
     }
@@ -924,43 +923,43 @@
     return E_NOINTERFACE;
 }
 
-static ULONG STDMETHODCALLTYPE d3d11_rasterizer_state_AddRef(ID3D11RasterizerState2 *iface)
-{
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+static ULONG STDMETHODCALLTYPE d3d11_rasterizer_state_AddRef(ID3D11RasterizerState1 *iface)
+{
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
     ULONG refcount = InterlockedIncrement(&state->refcount);
 
     TRACE("%p increasing refcount to %lu.\n", state, refcount);
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(state->device);
+        ID3D11Device2_AddRef(state->device);
         wined3d_rasterizer_state_incref(state->wined3d_state);
     }
 
     return refcount;
 }
 
-static ULONG STDMETHODCALLTYPE d3d11_rasterizer_state_Release(ID3D11RasterizerState2 *iface)
-{
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+static ULONG STDMETHODCALLTYPE d3d11_rasterizer_state_Release(ID3D11RasterizerState1 *iface)
+{
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
     ULONG refcount = InterlockedDecrement(&state->refcount);
 
     TRACE("%p decreasing refcount to %lu.\n", state, refcount);
 
     if (!refcount)
     {
-        ID3D11Device5 *device = state->device;
+        ID3D11Device2 *device = state->device;
         wined3d_rasterizer_state_decref(state->wined3d_state);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
 }
 
-static void STDMETHODCALLTYPE d3d11_rasterizer_state_GetDevice(ID3D11RasterizerState2 *iface,
+static void STDMETHODCALLTYPE d3d11_rasterizer_state_GetDevice(ID3D11RasterizerState1 *iface,
         ID3D11Device **device)
 {
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
 
     TRACE("iface %p, device %p.\n", iface, device);
 
@@ -968,67 +967,57 @@
     ID3D11Device_AddRef(*device);
 }
 
-static HRESULT STDMETHODCALLTYPE d3d11_rasterizer_state_GetPrivateData(ID3D11RasterizerState2 *iface,
+static HRESULT STDMETHODCALLTYPE d3d11_rasterizer_state_GetPrivateData(ID3D11RasterizerState1 *iface,
         REFGUID guid, UINT *data_size, void *data)
 {
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
 
     TRACE("iface %p, guid %s, data_size %p, data %p.\n", iface, debugstr_guid(guid), data_size, data);
 
     return d3d_get_private_data(&state->private_store, guid, data_size, data);
 }
 
-static HRESULT STDMETHODCALLTYPE d3d11_rasterizer_state_SetPrivateData(ID3D11RasterizerState2 *iface,
+static HRESULT STDMETHODCALLTYPE d3d11_rasterizer_state_SetPrivateData(ID3D11RasterizerState1 *iface,
         REFGUID guid, UINT data_size, const void *data)
 {
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
 
     TRACE("iface %p, guid %s, data_size %u, data %p.\n", iface, debugstr_guid(guid), data_size, data);
 
     return d3d_set_private_data(&state->private_store, guid, data_size, data);
 }
 
-static HRESULT STDMETHODCALLTYPE d3d11_rasterizer_state_SetPrivateDataInterface(ID3D11RasterizerState2 *iface,
+static HRESULT STDMETHODCALLTYPE d3d11_rasterizer_state_SetPrivateDataInterface(ID3D11RasterizerState1 *iface,
         REFGUID guid, const IUnknown *data)
 {
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
 
     TRACE("iface %p, guid %s, data %p.\n", iface, debugstr_guid(guid), data);
 
     return d3d_set_private_data_interface(&state->private_store, guid, data);
 }
 
-static void STDMETHODCALLTYPE d3d11_rasterizer_state_GetDesc(ID3D11RasterizerState2 *iface,
+static void STDMETHODCALLTYPE d3d11_rasterizer_state_GetDesc(ID3D11RasterizerState1 *iface,
         D3D11_RASTERIZER_DESC *desc)
 {
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
 
     TRACE("iface %p, desc %p.\n", iface, desc);
 
     memcpy(desc, &state->desc, sizeof(*desc));
 }
 
-static void STDMETHODCALLTYPE d3d11_rasterizer_state_GetDesc1(ID3D11RasterizerState2 *iface,
+static void STDMETHODCALLTYPE d3d11_rasterizer_state_GetDesc1(ID3D11RasterizerState1 *iface,
         D3D11_RASTERIZER_DESC1 *desc)
 {
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
+    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState1(iface);
 
     TRACE("iface %p, desc %p.\n", iface, desc);
 
-    memcpy(desc, &state->desc, sizeof(*desc));
-}
-
-static void STDMETHODCALLTYPE d3d11_rasterizer_state_GetDesc2(ID3D11RasterizerState2 *iface,
-        D3D11_RASTERIZER_DESC2 *desc)
-{
-    struct d3d_rasterizer_state *state = impl_from_ID3D11RasterizerState2(iface);
-
-    TRACE("iface %p, desc %p.\n", iface, desc);
-
     *desc = state->desc;
 }
 
-static const struct ID3D11RasterizerState2Vtbl d3d11_rasterizer_state_vtbl =
+static const struct ID3D11RasterizerState1Vtbl d3d11_rasterizer_state_vtbl =
 {
     /* IUnknown methods */
     d3d11_rasterizer_state_QueryInterface,
@@ -1043,8 +1032,6 @@
     d3d11_rasterizer_state_GetDesc,
     /* ID3D11RasterizerState1 methods */
     d3d11_rasterizer_state_GetDesc1,
-    /* ID3D11RasterizerState2 methods */
-    d3d11_rasterizer_state_GetDesc2,
 };
 
 /* ID3D10RasterizerState methods */
@@ -1063,7 +1050,7 @@
 
     TRACE("iface %p, riid %s, object %p.\n", iface, debugstr_guid(riid), object);
 
-    return d3d11_rasterizer_state_QueryInterface(&state->ID3D11RasterizerState2_iface, riid, object);
+    return d3d11_rasterizer_state_QueryInterface(&state->ID3D11RasterizerState1_iface, riid, object);
 }
 
 static ULONG STDMETHODCALLTYPE d3d10_rasterizer_state_AddRef(ID3D10RasterizerState *iface)
@@ -1072,7 +1059,7 @@
 
     TRACE("iface %p.\n", iface);
 
-    return d3d11_rasterizer_state_AddRef(&state->ID3D11RasterizerState2_iface);
+    return d3d11_rasterizer_state_AddRef(&state->ID3D11RasterizerState1_iface);
 }
 
 static ULONG STDMETHODCALLTYPE d3d10_rasterizer_state_Release(ID3D10RasterizerState *iface)
@@ -1081,7 +1068,7 @@
 
     TRACE("iface %p.\n", state);
 
-    return d3d11_rasterizer_state_Release(&state->ID3D11RasterizerState2_iface);
+    return d3d11_rasterizer_state_Release(&state->ID3D11RasterizerState1_iface);
 }
 
 /* ID3D10DeviceChild methods */
@@ -1092,7 +1079,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(state->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(state->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_rasterizer_state_GetPrivateData(ID3D10RasterizerState *iface,
@@ -1157,7 +1144,7 @@
 static void STDMETHODCALLTYPE d3d_rasterizer_state_wined3d_object_destroyed(void *parent)
 {
     struct d3d_rasterizer_state *state = parent;
-    struct d3d_device *device = impl_from_ID3D11Device5(state->device);
+    struct d3d_device *device = impl_from_ID3D11Device2(state->device);
 
     wine_rb_remove(&device->rasterizer_states, &state->entry);
     wined3d_private_store_cleanup(&state->private_store);
@@ -1180,12 +1167,12 @@
 }
 
 static HRESULT d3d_rasterizer_state_init(struct d3d_rasterizer_state *state, struct d3d_device *device,
-        const D3D11_RASTERIZER_DESC2 *desc)
+        const D3D11_RASTERIZER_DESC1 *desc)
 {
     struct wined3d_rasterizer_state_desc wined3d_desc;
     HRESULT hr;
 
-    state->ID3D11RasterizerState2_iface.lpVtbl = &d3d11_rasterizer_state_vtbl;
+    state->ID3D11RasterizerState1_iface.lpVtbl = &d3d11_rasterizer_state_vtbl;
     state->ID3D10RasterizerState_iface.lpVtbl = &d3d10_rasterizer_state_vtbl;
     state->refcount = 1;
     wined3d_private_store_init(&state->private_store);
@@ -1222,13 +1209,6 @@
             FIXME("Ignoring ForcedSampleCount %#x.\n", desc->ForcedSampleCount);
     }
 
-    if (desc->ConservativeRaster)
-    {
-        static unsigned int once;
-        if (!once++)
-            FIXME("Ignoring conservative rasterization.\n");
-    }
-
     /* We cannot fail after creating a wined3d_rasterizer_state object. It
      * would lead to double free. */
     if (FAILED(hr = wined3d_rasterizer_state_create(device->wined3d_device, &wined3d_desc,
@@ -1240,12 +1220,12 @@
         return hr;
     }
 
-    ID3D11Device5_AddRef(state->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(state->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
 
-HRESULT d3d_rasterizer_state_create(struct d3d_device *device, const D3D11_RASTERIZER_DESC2 *desc,
+HRESULT d3d_rasterizer_state_create(struct d3d_device *device, const D3D11_RASTERIZER_DESC1 *desc,
         struct d3d_rasterizer_state **state)
 {
     struct d3d_rasterizer_state *object;
@@ -1258,7 +1238,7 @@
         object = WINE_RB_ENTRY_VALUE(entry, struct d3d_rasterizer_state, entry);
 
         TRACE("Returning existing rasterizer state %p.\n", object);
-        ID3D11RasterizerState2_AddRef(&object->ID3D11RasterizerState2_iface);
+        ID3D11RasterizerState1_AddRef(&object->ID3D11RasterizerState1_iface);
         *state = object;
         wined3d_mutex_unlock();
 
@@ -1292,7 +1272,7 @@
         return NULL;
     assert(iface->lpVtbl == (ID3D11RasterizerStateVtbl *)&d3d11_rasterizer_state_vtbl);
 
-    return impl_from_ID3D11RasterizerState2((ID3D11RasterizerState2 *)iface);
+    return impl_from_ID3D11RasterizerState1((ID3D11RasterizerState1 *)iface);
 }
 
 struct d3d_rasterizer_state *unsafe_impl_from_ID3D10RasterizerState(ID3D10RasterizerState *iface)
@@ -1350,7 +1330,7 @@
 
     if (refcount == 1)
     {
-        ID3D11Device5_AddRef(state->device);
+        ID3D11Device2_AddRef(state->device);
         wined3d_sampler_incref(state->wined3d_sampler);
     }
 
@@ -1366,9 +1346,9 @@
 
     if (!refcount)
     {
-        ID3D11Device5 *device = state->device;
+        ID3D11Device2 *device = state->device;
         wined3d_sampler_decref(state->wined3d_sampler);
-        ID3D11Device5_Release(device);
+        ID3D11Device2_Release(device);
     }
 
     return refcount;
@@ -1485,7 +1465,7 @@
 
     TRACE("iface %p, device %p.\n", iface, device);
 
-    ID3D11Device5_QueryInterface(state->device, &IID_ID3D10Device, (void **)device);
+    ID3D11Device2_QueryInterface(state->device, &IID_ID3D10Device, (void **)device);
 }
 
 static HRESULT STDMETHODCALLTYPE d3d10_sampler_state_GetPrivateData(ID3D10SamplerState *iface,
@@ -1550,7 +1530,7 @@
 static void STDMETHODCALLTYPE d3d_sampler_wined3d_object_destroyed(void *parent)
 {
     struct d3d_sampler_state *state = parent;
-    struct d3d_device *device = impl_from_ID3D11Device5(state->device);
+    struct d3d_device *device = impl_from_ID3D11Device2(state->device);
 
     wine_rb_remove(&device->sampler_states, &state->entry);
     wined3d_private_store_cleanup(&state->private_store);
@@ -1588,9 +1568,9 @@
     return WINED3D_TEXF_POINT;
 }
 
-static enum wined3d_filter_reduction_mode wined3d_filter_reduction_mode_from_d3d11(enum D3D11_FILTER f)
-{
-    return (enum wined3d_filter_reduction_mode)D3D11_DECODE_FILTER_REDUCTION(f);
+static BOOL wined3d_texture_compare_from_d3d11(enum D3D11_FILTER f)
+{
+    return D3D11_DECODE_IS_COMPARISON_FILTER(f);
 }
 
 static HRESULT d3d_sampler_state_init(struct d3d_sampler_state *state, struct d3d_device *device,
@@ -1616,11 +1596,8 @@
     wined3d_desc.min_lod = desc->MinLOD;
     wined3d_desc.max_lod = max(desc->MinLOD, desc->MaxLOD);
     wined3d_desc.mip_base_level = 0;
-    if (D3D11_DECODE_IS_ANISOTROPIC_FILTER(desc->Filter) && desc->MaxAnisotropy)
-        wined3d_desc.max_anisotropy = desc->MaxAnisotropy;
-    else
-        wined3d_desc.max_anisotropy = 1;
-    wined3d_desc.reduction_mode = wined3d_filter_reduction_mode_from_d3d11(desc->Filter);
+    wined3d_desc.max_anisotropy = D3D11_DECODE_IS_ANISOTROPIC_FILTER(desc->Filter) ? desc->MaxAnisotropy : 1;
+    wined3d_desc.compare = wined3d_texture_compare_from_d3d11(desc->Filter);
     wined3d_desc.comparison_func = wined3d_cmp_func_from_d3d11(desc->ComparisonFunc);
     wined3d_desc.srgb_decode = TRUE;
 
@@ -1642,7 +1619,7 @@
         return hr;
     }
 
-    ID3D11Device5_AddRef(state->device = &device->ID3D11Device5_iface);
+    ID3D11Device2_AddRef(state->device = &device->ID3D11Device2_iface);
 
     return S_OK;
 }
