diff --git a/dlls/wined3d/context_vk.c b/dlls/wined3d/context_vk.c
--- a/dlls/wined3d/context_vk.c
+++ b/dlls/wined3d/context_vk.c
@@ -381,7 +381,7 @@
     return vk_memory;
 }
 
-struct wined3d_allocator_block *wined3d_context_vk_allocate_memory(struct wined3d_context_vk *context_vk,
+static struct wined3d_allocator_block *wined3d_context_vk_allocate_memory(struct wined3d_context_vk *context_vk,
         unsigned int memory_type, VkDeviceSize size, VkDeviceMemory *vk_memory)
 {
     struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
@@ -410,7 +410,7 @@
     return block;
 }
 
-void wined3d_context_vk_free_memory(struct wined3d_context_vk *context_vk, struct wined3d_allocator_block *block)
+static void wined3d_context_vk_free_memory(struct wined3d_context_vk *context_vk, struct wined3d_allocator_block *block)
 {
     struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
 
@@ -610,7 +610,7 @@
 BOOL wined3d_context_vk_create_image(struct wined3d_context_vk *context_vk, VkImageType vk_image_type,
         VkImageUsageFlags usage, VkFormat vk_format, unsigned int width, unsigned int height, unsigned int depth,
         unsigned int sample_count, unsigned int mip_levels, unsigned int layer_count, unsigned int flags,
-        const void *next, struct wined3d_image_vk *image)
+        struct wined3d_image_vk *image)
 {
     struct wined3d_adapter_vk *adapter_vk = wined3d_adapter_vk(context_vk->c.device->adapter);
     struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
@@ -621,7 +621,7 @@
     VkResult vr;
 
     create_info.sType = VK_STRUCTURE_TYPE_IMAGE_CREATE_INFO;
-    create_info.pNext = next;
+    create_info.pNext = NULL;
     create_info.flags = flags;
     create_info.imageType = vk_image_type;
     create_info.format = vk_format;
@@ -1089,60 +1089,6 @@
     o->command_buffer_id = command_buffer_id;
 }
 
-void wined3d_context_vk_destroy_vk_video_session(struct wined3d_context_vk *context_vk,
-        VkVideoSessionKHR vk_video_session, uint64_t command_buffer_id)
-{
-    struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
-    const struct wined3d_vk_info *vk_info = context_vk->vk_info;
-    struct wined3d_retired_object_vk *o;
-
-    /* The spec does not mention that the session may not be referenced by any
-     * active command buffers, but is probably an accidental omission.
-     * The proposal says "no longer used by any pending command buffers". */
-
-    if (context_vk->completed_command_buffer_id >= command_buffer_id)
-    {
-        VK_CALL(vkDestroyVideoSessionKHR(device_vk->vk_device, vk_video_session, NULL));
-        TRACE("Destroyed video session 0x%s.\n", wine_dbgstr_longlong(vk_video_session));
-        return;
-    }
-
-    if (!(o = wined3d_context_vk_get_retired_object_vk(context_vk)))
-    {
-        ERR("Leaking video session 0x%s.\n", wine_dbgstr_longlong(vk_video_session));
-        return;
-    }
-
-    o->type = WINED3D_RETIRED_VIDEO_SESSION_VK;
-    o->u.vk_video_session = vk_video_session;
-    o->command_buffer_id = command_buffer_id;
-}
-
-void wined3d_context_vk_destroy_vk_video_parameters(struct wined3d_context_vk *context_vk,
-        VkVideoSessionParametersKHR vk_video_parameters, uint64_t command_buffer_id)
-{
-    struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
-    const struct wined3d_vk_info *vk_info = context_vk->vk_info;
-    struct wined3d_retired_object_vk *o;
-
-    if (context_vk->completed_command_buffer_id >= command_buffer_id)
-    {
-        VK_CALL(vkDestroyVideoSessionParametersKHR(device_vk->vk_device, vk_video_parameters, NULL));
-        TRACE("Destroyed video parameters 0x%s.\n", wine_dbgstr_longlong(vk_video_parameters));
-        return;
-    }
-
-    if (!(o = wined3d_context_vk_get_retired_object_vk(context_vk)))
-    {
-        ERR("Leaking video parameters 0x%s.\n", wine_dbgstr_longlong(vk_video_parameters));
-        return;
-    }
-
-    o->type = WINED3D_RETIRED_VIDEO_PARAMETERS_VK;
-    o->u.vk_video_parameters = vk_video_parameters;
-    o->command_buffer_id = command_buffer_id;
-}
-
 void wined3d_context_vk_destroy_image(struct wined3d_context_vk *context_vk, struct wined3d_image_vk *image)
 {
     wined3d_context_vk_destroy_vk_image(context_vk, image->vk_image, image->command_buffer_id);
@@ -1223,116 +1169,6 @@
         free_command_buffer(context_vk, buffer);
 
     *buffer = context_vk->submitted.buffers[--context_vk->submitted.buffer_count];
-}
-
-bool wined3d_aux_command_pool_vk_get_buffer(struct wined3d_context_vk *context_vk,
-        struct wined3d_aux_command_pool_vk *pool, struct wined3d_aux_command_buffer_vk *buffer)
-{
-    VkCommandBufferBeginInfo begin_info = {.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_BEGIN_INFO};
-    struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
-    const struct wined3d_vk_info *vk_info = context_vk->vk_info;
-    VkResult vr;
-
-    if (pool->buffer_count)
-    {
-        *buffer = pool->buffers[--pool->buffer_count];
-    }
-    else
-    {
-        VkCommandBufferAllocateInfo command_buffer_info = {.sType = VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO};
-        VkSemaphoreCreateInfo semaphore_info = {.sType = VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO};
-
-        if ((vr = VK_CALL(vkCreateSemaphore(device_vk->vk_device, &semaphore_info, NULL, &buffer->signal_semaphore))) < 0)
-        {
-            ERR("Failed to create signal_semaphore, vr %s.\n", wined3d_debug_vkresult(vr));
-            return false;
-        }
-
-        if ((vr = VK_CALL(vkCreateSemaphore(device_vk->vk_device, &semaphore_info, NULL, &buffer->wait_semaphore))) < 0)
-        {
-            ERR("Failed to create wait_semaphore, vr %s.\n", wined3d_debug_vkresult(vr));
-            VK_CALL(vkDestroySemaphore(device_vk->vk_device, buffer->signal_semaphore, NULL));
-            return false;
-        }
-
-        command_buffer_info.commandPool = pool->vk_pool;
-        command_buffer_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
-        command_buffer_info.commandBufferCount = 1;
-        if ((vr = VK_CALL(vkAllocateCommandBuffers(device_vk->vk_device,
-                &command_buffer_info, &buffer->vk_command_buffer))) < 0)
-        {
-            WARN("Failed to allocate Vulkan command buffer, vr %s.\n", wined3d_debug_vkresult(vr));
-            VK_CALL(vkDestroySemaphore(device_vk->vk_device, buffer->signal_semaphore, NULL));
-            VK_CALL(vkDestroySemaphore(device_vk->vk_device, buffer->wait_semaphore, NULL));
-            return false;
-        }
-    }
-
-    begin_info.flags = VK_COMMAND_BUFFER_USAGE_ONE_TIME_SUBMIT_BIT;
-    if ((vr = VK_CALL(vkBeginCommandBuffer(buffer->vk_command_buffer, &begin_info))) < 0)
-    {
-        WARN("Failed to begin command buffer, vr %s.\n", wined3d_debug_vkresult(vr));
-        VK_CALL(vkFreeCommandBuffers(device_vk->vk_device, pool->vk_pool, 1, &buffer->vk_command_buffer));
-        return false;
-    }
-
-    return true;
-}
-
-void wined3d_aux_command_pool_vk_retire_buffer(
-        struct wined3d_context_vk *context_vk, struct wined3d_aux_command_pool_vk *pool,
-        const struct wined3d_aux_command_buffer_vk *buffer, uint64_t command_buffer_id)
-{
-    struct wined3d_retired_object_vk *o;
-
-    /* No point checking if it's complete. Auxiliary command buffers are
-     * always submitted before the main command buffer that depends on them. */
-
-    if (!(o = wined3d_context_vk_get_retired_object_vk(context_vk)))
-    {
-        ERR("Leaking auxiliary command buffer %p.\n", buffer->vk_command_buffer);
-        return;
-    }
-
-    o->type = WINED3D_RETIRED_AUX_COMMAND_BUFFER_VK;
-    o->u.aux_command_buffer.buffer = *buffer;
-    o->u.aux_command_buffer.pool = pool;
-    o->command_buffer_id = command_buffer_id;
-}
-
-static void wined3d_aux_command_pool_vk_complete_buffer(struct wined3d_context_vk *context_vk,
-        struct wined3d_aux_command_pool_vk *pool, const struct wined3d_aux_command_buffer_vk *buffer)
-{
-    struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
-    const struct wined3d_vk_info *vk_info = context_vk->vk_info;
-
-    if (!wined3d_array_reserve((void **)&pool->buffers, &pool->buffers_size,
-            pool->buffer_count + 1, sizeof(*pool->buffers)))
-    {
-        VK_CALL(vkFreeCommandBuffers(device_vk->vk_device, pool->vk_pool, 1, &buffer->vk_command_buffer));
-        return;
-    }
-
-    pool->buffers[pool->buffer_count++] = *buffer;
-}
-
-static void wined3d_aux_command_pool_vk_cleanup(struct wined3d_context_vk *context_vk,
-        struct wined3d_aux_command_pool_vk *pool)
-{
-    struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
-    const struct wined3d_vk_info *vk_info = context_vk->vk_info;
-
-    for (unsigned int i = 0; i < pool->buffer_count; ++i)
-    {
-        struct wined3d_aux_command_buffer_vk *buffer = &pool->buffers[i];
-
-        VK_CALL(vkDestroySemaphore(device_vk->vk_device, buffer->wait_semaphore, NULL));
-        VK_CALL(vkDestroySemaphore(device_vk->vk_device, buffer->signal_semaphore, NULL));
-        VK_CALL(vkFreeCommandBuffers(device_vk->vk_device,
-                pool->vk_pool, 1, &buffer->vk_command_buffer));
-    }
-    VK_CALL(vkDestroyCommandPool(device_vk->vk_device, pool->vk_pool, NULL));
-    free(pool->buffers);
 }
 
 static void wined3d_context_vk_cleanup_resources(struct wined3d_context_vk *context_vk, VkFence vk_fence)
@@ -1439,21 +1275,6 @@
             case WINED3D_RETIRED_PIPELINE_VK:
                 VK_CALL(vkDestroyPipeline(device_vk->vk_device, o->u.vk_pipeline, NULL));
                 TRACE("Destroyed pipeline 0x%s.\n", wine_dbgstr_longlong(o->u.vk_pipeline));
-                break;
-
-            case WINED3D_RETIRED_VIDEO_SESSION_VK:
-                VK_CALL(vkDestroyVideoSessionKHR(device_vk->vk_device, o->u.vk_video_session, NULL));
-                TRACE("Destroyed video session 0x%s.\n", wine_dbgstr_longlong(o->u.vk_video_session));
-                break;
-
-            case WINED3D_RETIRED_VIDEO_PARAMETERS_VK:
-                VK_CALL(vkDestroyVideoSessionParametersKHR(device_vk->vk_device, o->u.vk_video_parameters, NULL));
-                TRACE("Destroyed video parameters 0x%s.\n", wine_dbgstr_longlong(o->u.vk_video_parameters));
-                break;
-
-            case WINED3D_RETIRED_AUX_COMMAND_BUFFER_VK:
-                wined3d_aux_command_pool_vk_complete_buffer(context_vk,
-                        o->u.aux_command_buffer.pool, &o->u.aux_command_buffer.buffer);
                 break;
 
             default:
@@ -1933,7 +1754,6 @@
      * this needs to happen after all command buffers are freed, because
      * vkFreeCommandBuffers() requires a valid pool handle. */
     VK_CALL(vkDestroyCommandPool(device_vk->vk_device, context_vk->vk_command_pool, NULL));
-    wined3d_aux_command_pool_vk_cleanup(context_vk, &context_vk->decode_pool);
     wined3d_context_vk_destroy_query_pools(context_vk, &context_vk->free_occlusion_query_pools);
     wined3d_context_vk_destroy_query_pools(context_vk, &context_vk->free_timestamp_query_pools);
     wined3d_context_vk_destroy_query_pools(context_vk, &context_vk->free_pipeline_statistics_query_pools);
@@ -2078,11 +1898,11 @@
         unsigned int signal_semaphore_count, const VkSemaphore *signal_semaphores)
 {
     struct wined3d_device_vk *device_vk = wined3d_device_vk(context_vk->c.device);
-    VkSubmitInfo submit_info = {.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO};
     const struct wined3d_vk_info *vk_info = context_vk->vk_info;
     struct wined3d_query_pool_vk *pool_vk, *pool_vk_next;
     struct wined3d_command_buffer_vk *buffer;
     struct wined3d_query_vk *query_vk;
+    VkSubmitInfo submit_info;
     VkResult vr;
 
     TRACE("context_vk %p, wait_semaphore_count %u, wait_semaphores %p, wait_stages %p,"
@@ -2091,89 +1911,63 @@
             signal_semaphore_count, signal_semaphores);
 
     buffer = &context_vk->current_command_buffer;
-    if (!buffer->vk_command_buffer && !signal_semaphore_count
-            && !wait_semaphore_count && !context_vk->wait_semaphore_count)
+    if (!buffer->vk_command_buffer)
         return;
 
-    if (buffer->vk_command_buffer)
-    {
-        TRACE("Submitting command buffer %p with id 0x%s.\n",
-                buffer->vk_command_buffer, wine_dbgstr_longlong(buffer->id));
-
-        wined3d_context_vk_end_current_render_pass(context_vk);
-
-        LIST_FOR_EACH_ENTRY_SAFE(pool_vk, pool_vk_next, &context_vk->completed_query_pools,
-                struct wined3d_query_pool_vk, completed_entry)
-        {
-            list_remove(&pool_vk->completed_entry);
-            list_init(&pool_vk->completed_entry);
-
-            wined3d_context_vk_reset_completed_queries(context_vk, pool_vk, buffer);
-        }
-
-        LIST_FOR_EACH_ENTRY(query_vk, &context_vk->active_queries, struct wined3d_query_vk, entry)
-            wined3d_query_vk_suspend(query_vk, context_vk);
-
-        context_vk->graphics.vk_pipeline = VK_NULL_HANDLE;
-        context_vk->update_compute_pipeline = 1;
-        context_vk->update_stream_output = 1;
-        context_vk->c.update_shader_resource_bindings = 1;
-        context_vk->c.update_compute_shader_resource_bindings = 1;
-        context_vk->c.update_unordered_access_view_bindings = 1;
-        context_vk->c.update_compute_unordered_access_view_bindings = 1;
-        context_vk->c.update_primitive_type = 1;
-        context_vk->c.update_patch_vertex_count = 1;
-        context_vk->c.update_multisample_state = 1;
-        context_invalidate_state(&context_vk->c, STATE_STREAMSRC);
-        context_invalidate_state(&context_vk->c, STATE_INDEXBUFFER);
-        context_invalidate_state(&context_vk->c, STATE_BLEND);
-        context_invalidate_state(&context_vk->c, STATE_BLEND_FACTOR);
-        context_invalidate_state(&context_vk->c, STATE_DEPTH_STENCIL);
-        context_invalidate_state(&context_vk->c, STATE_RASTERIZER);
-        context_invalidate_state(&context_vk->c, STATE_STENCIL_REF);
-        context_invalidate_state(&context_vk->c, STATE_VIEWPORT);
-        context_invalidate_state(&context_vk->c, STATE_SCISSORRECT);
-
-        VK_CALL(vkEndCommandBuffer(buffer->vk_command_buffer));
-
-        VK_CALL(vkResetFences(device_vk->vk_device, 1, &buffer->vk_fence));
-    }
-
-    if (wait_semaphore_count)
-    {
-        wined3d_array_reserve((void **)&context_vk->wait_semaphores, &context_vk->wait_semaphores_size,
-                context_vk->wait_semaphore_count + wait_semaphore_count, sizeof(*context_vk->wait_semaphores));
-        memcpy(context_vk->wait_semaphores + context_vk->wait_semaphore_count,
-                wait_semaphores, wait_semaphore_count * sizeof(VkSemaphore));
-
-        wined3d_array_reserve((void **)&context_vk->wait_stages, &context_vk->wait_stages_size,
-                context_vk->wait_semaphore_count + wait_semaphore_count, sizeof(*context_vk->wait_stages));
-        memcpy(context_vk->wait_stages + context_vk->wait_semaphore_count,
-                wait_stages, wait_semaphore_count * sizeof(VkPipelineStageFlags));
-
-        context_vk->wait_semaphore_count += wait_semaphore_count;
-    }
-
-    submit_info.waitSemaphoreCount = context_vk->wait_semaphore_count;
-    submit_info.pWaitSemaphores = context_vk->wait_semaphores;
-    submit_info.pWaitDstStageMask = context_vk->wait_stages;
-    if (buffer->vk_command_buffer)
-    {
-        submit_info.commandBufferCount = 1;
-        submit_info.pCommandBuffers = &buffer->vk_command_buffer;
-    }
+    TRACE("Submitting command buffer %p with id 0x%s.\n",
+            buffer->vk_command_buffer, wine_dbgstr_longlong(buffer->id));
+
+    wined3d_context_vk_end_current_render_pass(context_vk);
+
+    LIST_FOR_EACH_ENTRY_SAFE(pool_vk, pool_vk_next, &context_vk->completed_query_pools,
+            struct wined3d_query_pool_vk, completed_entry)
+    {
+        list_remove(&pool_vk->completed_entry);
+        list_init(&pool_vk->completed_entry);
+
+        wined3d_context_vk_reset_completed_queries(context_vk, pool_vk, buffer);
+    }
+
+    LIST_FOR_EACH_ENTRY(query_vk, &context_vk->active_queries, struct wined3d_query_vk, entry)
+        wined3d_query_vk_suspend(query_vk, context_vk);
+
+    context_vk->graphics.vk_pipeline = VK_NULL_HANDLE;
+    context_vk->update_compute_pipeline = 1;
+    context_vk->update_stream_output = 1;
+    context_vk->c.update_shader_resource_bindings = 1;
+    context_vk->c.update_compute_shader_resource_bindings = 1;
+    context_vk->c.update_unordered_access_view_bindings = 1;
+    context_vk->c.update_compute_unordered_access_view_bindings = 1;
+    context_vk->c.update_primitive_type = 1;
+    context_vk->c.update_patch_vertex_count = 1;
+    context_vk->c.update_multisample_state = 1;
+    context_invalidate_state(&context_vk->c, STATE_STREAMSRC);
+    context_invalidate_state(&context_vk->c, STATE_INDEXBUFFER);
+    context_invalidate_state(&context_vk->c, STATE_BLEND);
+    context_invalidate_state(&context_vk->c, STATE_BLEND_FACTOR);
+    context_invalidate_state(&context_vk->c, STATE_DEPTH_STENCIL);
+    context_invalidate_state(&context_vk->c, STATE_RASTERIZER);
+    context_invalidate_state(&context_vk->c, STATE_STENCIL_REF);
+    context_invalidate_state(&context_vk->c, STATE_VIEWPORT);
+    context_invalidate_state(&context_vk->c, STATE_SCISSORRECT);
+
+    VK_CALL(vkEndCommandBuffer(buffer->vk_command_buffer));
+
+    VK_CALL(vkResetFences(device_vk->vk_device, 1, &buffer->vk_fence));
+
+    submit_info.sType = VK_STRUCTURE_TYPE_SUBMIT_INFO;
+    submit_info.pNext = NULL;
+    submit_info.waitSemaphoreCount = wait_semaphore_count;
+    submit_info.pWaitSemaphores = wait_semaphores;
+    submit_info.pWaitDstStageMask = wait_stages;
+    submit_info.commandBufferCount = 1;
+    submit_info.pCommandBuffers = &buffer->vk_command_buffer;
     submit_info.signalSemaphoreCount = signal_semaphore_count;
     submit_info.pSignalSemaphores = signal_semaphores;
 
-    if ((vr = VK_CALL(vkQueueSubmit(device_vk->graphics_queue.vk_queue, 1, &submit_info,
-            buffer->vk_command_buffer ? buffer->vk_fence : VK_NULL_HANDLE))) < 0)
+    if ((vr = VK_CALL(vkQueueSubmit(device_vk->vk_queue, 1, &submit_info, buffer->vk_fence))) < 0)
         ERR("Failed to submit command buffer %p, vr %s.\n",
                 buffer->vk_command_buffer, wined3d_debug_vkresult(vr));
-
-    context_vk->wait_semaphore_count = 0;
-
-    if (!buffer->vk_command_buffer)
-        return;
 
     if (!wined3d_array_reserve((void **)&context_vk->submitted.buffers, &context_vk->submitted.buffers_size,
             context_vk->submitted.buffer_count + 1, sizeof(*context_vk->submitted.buffers)))
@@ -4354,26 +4148,13 @@
     return vk_command_buffer;
 }
 
-static VkCommandPool create_command_pool(struct wined3d_device_vk *device_vk,
-        const struct wined3d_vk_info *vk_info, uint32_t queue_family_index)
-{
-    VkCommandPoolCreateInfo command_pool_info = {.sType = VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO};
-    VkCommandPool pool;
-    VkResult vr;
-
-    command_pool_info.flags = VK_COMMAND_POOL_CREATE_TRANSIENT_BIT | VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT;
-    command_pool_info.queueFamilyIndex = queue_family_index;
-    if ((vr = VK_CALL(vkCreateCommandPool(device_vk->vk_device, &command_pool_info, NULL, &pool))) == VK_SUCCESS)
-        return pool;
-    ERR("Failed to create Vulkan command pool, vr %s.\n", wined3d_debug_vkresult(vr));
-    return VK_NULL_HANDLE;
-}
-
 HRESULT wined3d_context_vk_init(struct wined3d_context_vk *context_vk, struct wined3d_swapchain *swapchain)
 {
+    VkCommandPoolCreateInfo command_pool_info;
     const struct wined3d_vk_info *vk_info;
     struct wined3d_adapter_vk *adapter_vk;
     struct wined3d_device_vk *device_vk;
+    VkResult vr;
 
     TRACE("context_vk %p, swapchain %p.\n", context_vk, swapchain);
 
@@ -4383,22 +4164,18 @@
     adapter_vk = wined3d_adapter_vk(device_vk->d.adapter);
     context_vk->vk_info = vk_info = &adapter_vk->vk_info;
 
-    if (!(context_vk->vk_command_pool = create_command_pool(device_vk,
-            vk_info, device_vk->graphics_queue.vk_queue_family_index)))
-    {
+    command_pool_info.sType = VK_STRUCTURE_TYPE_COMMAND_POOL_CREATE_INFO;
+    command_pool_info.pNext = NULL;
+    command_pool_info.flags = VK_COMMAND_POOL_CREATE_TRANSIENT_BIT | VK_COMMAND_POOL_CREATE_RESET_COMMAND_BUFFER_BIT;
+    command_pool_info.queueFamilyIndex = device_vk->vk_queue_family_index;
+    if ((vr = VK_CALL(vkCreateCommandPool(device_vk->vk_device,
+            &command_pool_info, NULL, &context_vk->vk_command_pool))) < 0)
+    {
+        ERR("Failed to create Vulkan command pool, vr %s.\n", wined3d_debug_vkresult(vr));
         wined3d_context_cleanup(&context_vk->c);
         return E_FAIL;
     }
     context_vk->current_command_buffer.id = 1;
-
-    if (device_vk->decode_queue.vk_queue
-            && !(context_vk->decode_pool.vk_pool = create_command_pool(device_vk,
-                    vk_info, device_vk->decode_queue.vk_queue_family_index)))
-    {
-        VK_CALL(vkDestroyCommandPool(device_vk->vk_device, context_vk->vk_command_pool, NULL));
-        wined3d_context_cleanup(&context_vk->c);
-        return E_FAIL;
-    }
 
     wined3d_context_vk_init_graphics_pipeline_key(context_vk);
 
