diff --git a/dlls/winegstreamer/wg_transform.c b/dlls/winegstreamer/wg_transform.c
--- a/dlls/winegstreamer/wg_transform.c
+++ b/dlls/winegstreamer/wg_transform.c
@@ -65,10 +65,7 @@
 #define GST_IS_WG_STEPPER_CLASS(klass) (G_TYPE_CHECK_CLASS_TYPE((klass),GST_TYPE_WG_STEPPER))
 
 G_DEFINE_TYPE (WgStepper, wg_stepper, GST_TYPE_ELEMENT);
-gboolean gst_element_register_winegstreamerstepper(GstPlugin *plugin)
-{
-    return gst_element_register(plugin, "winegstreamerstepper", GST_RANK_NONE, GST_TYPE_WG_STEPPER);
-}
+GST_ELEMENT_REGISTER_DEFINE(winegstreamerstepper, "winegstreamerstepper", GST_RANK_NONE, GST_TYPE_WG_STEPPER);
 
 static bool wg_stepper_step(WgStepper *stepper);
 static void wg_stepper_flush(WgStepper *stepper);
@@ -104,7 +101,7 @@
     return (struct wg_transform *)(ULONG_PTR)trans;
 }
 
-static void align_video_info_planes(MFVideoInfo *video_info, gsize plane_align,
+static void align_video_info_planes(MFVideoInfo *video_info, gsize plane_align, guint stride,
         GstVideoInfo *info, GstVideoAlignment *align)
 {
     bool fix_nv12 = !plane_align && info->finfo->format == GST_VIDEO_FORMAT_NV12 && (info->width & 3) && (info->width & 3) != 3;
@@ -115,12 +112,33 @@
     align->padding_right = ((plane_align + 1) - (info->width & plane_align)) & plane_align;
     align->padding_bottom = ((plane_align + 1) - (info->height & plane_align)) & plane_align;
 
-    if (!is_mf_video_area_empty(aperture))
+    if (!is_mf_video_area_empty(aperture) && !plane_align)
     {
         align->padding_right = max(align->padding_right, video_info->dwWidth - aperture->OffsetX.value - aperture->Area.cx);
         align->padding_bottom = max(align->padding_bottom, video_info->dwHeight - aperture->OffsetY.value - aperture->Area.cy);
         align->padding_top = aperture->OffsetX.value;
         align->padding_left = aperture->OffsetY.value;
+    }
+
+    if (stride)
+    {
+        /* The MF sample has a 2D buffer. Set padding_right to match its stride. */
+        guint width = align->padding_left + info->width + align->padding_right;
+        const GstVideoFormatInfo *finfo = info->finfo;
+        gint comp[GST_VIDEO_MAX_COMPONENTS];
+        gint pixel_stride;
+
+        gst_video_format_info_component(finfo, 0, comp);
+        pixel_stride = finfo->pixel_stride[comp[0]];
+
+        if (stride % pixel_stride)
+            GST_ERROR("Stride %u not aligned to pixel size", stride);
+        stride /= pixel_stride;
+
+        if (stride < width)
+            GST_ERROR("Invalid stride %u", stride);
+        else
+            align->padding_right += stride - width;
     }
 
     if (video_info->VideoFlags & MFVideoFlag_BottomUpLinearRep)
@@ -217,7 +235,7 @@
     pool_class->alloc_buffer = wg_video_buffer_pool_alloc_buffer;
 }
 
-static WgVideoBufferPool *wg_video_buffer_pool_create(GstCaps *caps, gsize plane_align,
+static WgVideoBufferPool *wg_video_buffer_pool_create(GstCaps *caps, gsize plane_align, gsize output_plane_stride,
         GstAllocator *allocator, MFVideoInfo *video_info, GstVideoAlignment *align)
 {
     WgVideoBufferPool *pool;
@@ -229,7 +247,7 @@
 
     gst_video_info_from_caps(&pool->info, caps);
     max_size = pool->info.size;
-    align_video_info_planes(video_info, plane_align, &pool->info, align);
+    align_video_info_planes(video_info, plane_align, output_plane_stride, &pool->info, align);
     /* GStreamer assumes NV12 pools must accommodate a stride alignment of 4, but we use 2 */
     max_size = max(max_size, pool->info.size);
 
@@ -313,7 +331,7 @@
         return false;
 
     if (!(pool = wg_video_buffer_pool_create(caps, transform->attrs.output_plane_align,
-            transform->allocator, &transform->output_info, &align)))
+            transform->attrs.output_plane_stride, transform->allocator, &transform->output_info, &align)))
         return false;
 
     if ((params = gst_structure_new("video-meta",
@@ -520,18 +538,85 @@
     return parsed_caps;
 }
 
+static GstBuffer *caps_get_buffer(const GstCaps *caps, const char *name, UINT32 *buffer_size)
+{
+    const GstStructure *structure = gst_caps_get_structure(caps, 0);
+    const GValue *buffer_value;
+
+    if ((buffer_value = gst_structure_get_value(structure, name)))
+    {
+        GstBuffer *buffer = gst_value_get_buffer(buffer_value);
+        *buffer_size = gst_buffer_get_size(buffer);
+        return buffer;
+    }
+
+    *buffer_size = 0;
+    return NULL;
+}
+
+static void push_vorbis_headers(struct wg_transform *transform)
+{
+    const uint8_t *ptr, *beg, *end;
+    GstBuffer *codec_data, *hdr;
+    UINT32 codec_data_size;
+    GstBufferMapInfo info;
+    int i, count, len;
+
+    if (!(codec_data = caps_get_buffer(transform->input_caps, "codec_data",
+            &codec_data_size)) || !codec_data_size) return;
+    gst_buffer_map(codec_data, &info, GST_MAP_READ);
+    ptr = info.data;
+    end = ptr + info.size;
+
+    for (len = 0, i = 0, count = *ptr++; ptr < end && i < count; i++)
+    {
+        while (ptr < end && *ptr++ == 0xff) len += 0xff;
+        len += ptr[-1];
+        GST_DEBUG("buffer %d: %u bytes", i, len);
+    }
+    if (len > end - ptr) goto failed;
+    beg = ptr;
+    ptr = info.data;
+
+    GST_DEBUG("%u stream headers, total length=%u bytes", count + 1, codec_data_size);
+    for (len = 0, i = 0, count = *ptr++; ptr < end && i < count; i++, len = 0)
+    {
+        while (ptr < end && *ptr++ == 0xff) len += 0xff;
+        len += ptr[-1];
+
+        if (!(hdr = gst_buffer_new_memdup(beg, len))) break;
+        GST_DEBUG("buffer %d: %u bytes", i, len);
+        GST_BUFFER_FLAG_SET(hdr, GST_BUFFER_FLAG_HEADER);
+        GST_MEMDUMP("data", beg, len);
+        gst_pad_push(transform->my_src, hdr);
+        beg += len;
+    }
+
+    if ((hdr = gst_buffer_new_memdup(beg, end - beg)))
+    {
+        GST_DEBUG("buffer %d: %zu bytes", i, end - beg);
+        GST_MEMDUMP("data", beg, end - beg);
+        GST_BUFFER_FLAG_SET(hdr, GST_BUFFER_FLAG_HEADER);
+        gst_pad_push(transform->my_src, hdr);
+    }
+
+failed:
+    gst_buffer_unmap(codec_data, &info);
+}
+
+
 static bool transform_create_decoder_elements(struct wg_transform *transform,
         const gchar *input_mime, const gchar *output_mime, GstElement **first, GstElement **last)
 {
     GstCaps *parsed_caps = NULL, *sink_caps = NULL;
-    GstElement *element, *capsfilter;
-    const char *shortname = NULL;
-    GstElementFactory *factory;
+    GstElement *element;
     bool ret = false;
-    char *str;
 
     if (!strcmp(input_mime, "audio/x-raw") || !strcmp(input_mime, "video/x-raw"))
+    {
+        transform->attrs.input_queue_length = 16;
         return true;
+    }
 
     if (!(parsed_caps = transform_get_parsed_caps(transform->input_caps, input_mime)))
         return false;
@@ -563,38 +648,8 @@
         parsed_caps = gst_caps_ref(transform->input_caps);
     }
 
-    if (!(element = find_element(GST_ELEMENT_FACTORY_TYPE_DECODER, parsed_caps, sink_caps)))
-        goto done;
-
-    /* h264parse currently has a bug where it will send an avc caps without codec_data when it
-     * has received an SPS but not PPS. As a result, when a drain request is made, the caps is fixated
-     * to avc but no codec_data is provided to the decoder. This results in libav rejecting every
-     * packet it receives.
-     * As a workaround, we need to insert a capsfilter for avdec_h264 in order for it to use
-     * the byte-stream stream-format.
-     */
-    if ((factory = gst_element_get_factory(element)) &&
-            (shortname = gst_plugin_feature_get_name(GST_PLUGIN_FEATURE(factory))) &&
-            !strcmp(shortname, "avdec_h264"))
-    {
-        if (!(capsfilter = create_element("capsfilter", "good")) ||
-                !append_element(transform->container, capsfilter, first, last))
-        {
-            g_object_unref(element);
-            goto done;
-        }
-
-        gst_caps_set_simple(parsed_caps, "stream-format", G_TYPE_STRING, "byte-stream", NULL);
-        gst_caps_set_simple(parsed_caps, "alignment", G_TYPE_STRING, "au", NULL);
-
-        if ((str = gst_caps_to_string(parsed_caps)))
-        {
-            gst_util_set_object_arg(G_OBJECT(capsfilter), "caps", str);
-            free(str);
-        }
-    }
-
-    if (!append_element(transform->container, element, first, last))
+    if (!(element = find_element(GST_ELEMENT_FACTORY_TYPE_DECODER, parsed_caps, sink_caps))
+            || !append_element(transform->container, element, first, last))
         goto done;
 
     set_max_threads(element);
@@ -660,6 +715,7 @@
 NTSTATUS wg_transform_create(void *args)
 {
     struct wg_transform_create_params *params = args;
+    struct wg_media_type input_type;
     GstElement *first = NULL, *last = NULL;
     NTSTATUS status = STATUS_UNSUCCESSFUL;
     const gchar *input_mime, *output_mime;
@@ -681,18 +737,46 @@
         goto out;
     transform->attrs = params->attrs;
 
-    if (!(transform->input_caps = caps_from_media_type(&params->input_type)))
+    memcpy(&input_type, &params->input_type, sizeof(input_type));
+
+    if (IsEqualGUID(&input_type.major, &MFMediaType_Audio))
+    {
+        size_t data_size = input_type.u.audio->cbSize + sizeof(WAVEFORMATEX) - sizeof(HEAACWAVEINFO);
+
+        /* If an mfsrcsnk hack appended transcoded audio info to the user data, then restore it.
+         * This happens if the game depends on the input format belonging to a specific set of formats. */
+        if (input_type.u.audio->wFormatTag == WAVE_FORMAT_MPEG_HEAAC
+                && data_size >= sizeof(WAVEFORMATEXTENSIBLE))
+        {
+            const HEAACWAVEFORMAT *hwf = (HEAACWAVEFORMAT *)input_type.u.audio;
+            WAVEFORMATEXTENSIBLE audio;
+
+            memcpy(&audio, &hwf->pbAudioSpecificConfig[data_size - sizeof(audio)], sizeof(audio));
+            if (audio.Format.wFormatTag == WAVE_FORMAT_EXTENSIBLE
+                    && IsEqualGUID(&audio.SubFormat, &MFAudioFormat_Vorbis))
+            {
+                memmove((WAVEFORMATEXTENSIBLE *)input_type.u.audio + 1, hwf->pbAudioSpecificConfig,
+                        data_size - sizeof(audio));
+                memcpy(input_type.u.audio, &audio, sizeof(audio));
+            }
+        }
+    }
+
+    if (!(transform->input_caps = caps_from_media_type(&input_type)))
         goto out;
     GST_INFO("transform %p input caps %"GST_PTR_FORMAT, transform, transform->input_caps);
     input_mime = gst_structure_get_name(gst_caps_get_structure(transform->input_caps, 0));
 
+    if (!strcmp(input_mime, "video/x-h264"))
+        touch_h264_used_tag();
+
     if (!(transform->output_caps = caps_from_media_type(&params->output_type)))
         goto out;
     GST_INFO("transform %p output caps %"GST_PTR_FORMAT, transform, transform->output_caps);
     output_mime = gst_structure_get_name(gst_caps_get_structure(transform->output_caps, 0));
 
-    if (IsEqualGUID(&params->input_type.major, &MFMediaType_Video))
-        transform->input_info = params->input_type.u.video->videoInfo;
+    if (IsEqualGUID(&input_type.major, &MFMediaType_Video))
+        transform->input_info = input_type.u.video->videoInfo;
     if (IsEqualGUID(&params->output_type.major, &MFMediaType_Video))
         transform->output_info = params->output_type.u.video->videoInfo;
 
@@ -746,6 +830,20 @@
     if (!(event = gst_event_new_caps(transform->input_caps))
             || !push_event(transform->my_src, event))
         goto out;
+
+    /* Check that the caps event have been accepted */
+    if (!strcmp(input_mime, "video/x-h264"))
+    {
+        GstPad *peer;
+        if (!(peer = gst_pad_get_peer(transform->my_src)))
+            goto out;
+        else if (!gst_pad_has_current_caps(peer))
+        {
+            gst_object_unref(peer);
+            goto out;
+        }
+        gst_object_unref(peer);
+    }
 
     /* We need to use GST_FORMAT_TIME here because it's the only format
      * some elements such avdec_wmav2 correctly support. */
@@ -755,6 +853,9 @@
     if (!(event = gst_event_new_segment(&transform->segment))
             || !push_event(transform->my_src, event))
         goto out;
+
+    if (!strcmp(input_mime, "audio/x-vorbis"))
+        push_vorbis_headers(transform);
 
     GST_INFO("Created winegstreamer transform %p.", transform);
     params->transform = (wg_transform_t)(ULONG_PTR)transform;
@@ -896,7 +997,7 @@
     }
 
     if (!(buffer = gst_buffer_new_wrapped_full(GST_MEMORY_FLAG_READONLY, wg_sample_data(sample), sample->max_size,
-            0, sample->size, sample, wg_sample_free_notify)))
+            0, sample->stride ? sample->max_size : sample->size, sample, wg_sample_free_notify)))
     {
         GST_ERROR("Failed to allocate input buffer");
         return STATUS_NO_MEMORY;
@@ -911,7 +1012,7 @@
     if (!strcmp(input_mime, "video/x-raw") && gst_video_info_from_caps(&video_info, transform->input_caps))
     {
         GstVideoAlignment align;
-        align_video_info_planes(&transform->input_info, 0, &video_info, &align);
+        align_video_info_planes(&transform->input_info, 0, sample->stride, &video_info, &align);
         buffer_add_video_meta(buffer, &video_info);
     }
 
@@ -949,8 +1050,8 @@
     return STATUS_SUCCESS;
 }
 
-static NTSTATUS copy_video_buffer(GstBuffer *buffer, GstVideoInfo *src_video_info,
-        GstVideoInfo *dst_video_info, struct wg_sample *sample, gsize *total_size)
+static NTSTATUS copy_video_buffer(GstBuffer *buffer, const GstVideoInfo *src_video_info,
+        const GstVideoInfo *dst_video_info, struct wg_sample *sample, gsize *total_size, GstBuffer **ret_buffer)
 {
     NTSTATUS status = STATUS_UNSUCCESSFUL;
     GstVideoFrame src_frame, dst_frame;
@@ -988,7 +1089,10 @@
         gst_video_frame_unmap(&src_frame);
     }
 
-    gst_buffer_unref(dst_buffer);
+    if (status == STATUS_SUCCESS)
+        *ret_buffer = dst_buffer;
+    else
+        gst_buffer_unref(dst_buffer);
     return status;
 }
 
@@ -1081,17 +1185,70 @@
     return needs_copy;
 }
 
+enum fill_action
+{
+    FILL_RIGHT  = 1,
+    FILL_BOTTOM = 2,
+};
+
+static void fill_frame_padded_bits(GstBuffer *buffer, const GstVideoAlignment *align, const GstVideoInfo *info,
+        enum fill_action action)
+{
+    guint i, j, plane, padded_height, width, height, stride, pixel_stride, padding_bottom = align->padding_bottom;
+    GstVideoFrame frame;
+
+    if (!padding_bottom) action &= ~FILL_BOTTOM;
+    if (!align->padding_right) action &= ~FILL_RIGHT;
+
+    if (!action || !gst_video_frame_map(&frame, info, buffer, GST_MAP_WRITE)) return;
+
+    /* Windows uses the data in the last scanline for its bottom padding, and the last pixel
+     * in a row for right padding. GStreamer can do this, but it requires cropping first, then
+     * edge replication using videobox, so it has a larger performance cost than this hack. */
+    for (plane = 0; plane < GST_VIDEO_FRAME_N_PLANES(&frame); plane++)
+    {
+        gint comp[GST_VIDEO_MAX_COMPONENTS];
+
+        gst_video_format_info_component(frame.info.finfo, plane, comp);
+        height = GST_VIDEO_FRAME_COMP_HEIGHT(&frame, comp[0]);
+        stride = GST_VIDEO_FRAME_PLANE_STRIDE(&frame, plane);
+
+        if (action & FILL_RIGHT)
+        {
+            guint8 *data = GST_VIDEO_FRAME_PLANE_DATA(&frame, plane);
+            pixel_stride = GST_VIDEO_FRAME_COMP_PSTRIDE(&frame, plane);
+            width = GST_VIDEO_FRAME_COMP_WIDTH(&frame, comp[0]) * pixel_stride;
+            data += width;
+            for (i = 0; i < height; i++)
+                for (j = 0; j < stride - width; j += pixel_stride)
+                    memcpy(data + i * stride + j, data + i * stride - pixel_stride, pixel_stride);
+        }
+
+        if (action & FILL_BOTTOM)
+        {
+            guint8 *data = GST_VIDEO_FRAME_PLANE_DATA(&frame, plane);
+            padded_height = GST_VIDEO_FORMAT_INFO_SCALE_HEIGHT(frame.info.finfo, comp[0], info->height + padding_bottom);
+            data += height * stride;
+            for (i = 0; i < padded_height - height; i++) memcpy(data + i * stride, data - stride, stride);
+        }
+    }
+
+    gst_video_frame_unmap(&frame);
+}
+
 static NTSTATUS read_transform_output_video(struct wg_sample *sample, GstBuffer *buffer,
-        GstVideoInfo *src_video_info, GstVideoInfo *dst_video_info)
-{
+        const GstVideoInfo *src_video_info, const GstVideoInfo *dst_video_info, const GstVideoAlignment *align)
+{
+    GstBuffer *dst_buffer = NULL;
     gsize total_size;
     NTSTATUS status;
     bool needs_copy;
+    const char *sgi;
 
     if (!(needs_copy = sample_needs_buffer_copy(sample, buffer, &total_size)))
         status = STATUS_SUCCESS;
     else
-        status = copy_video_buffer(buffer, src_video_info, dst_video_info, sample, &total_size);
+        status = copy_video_buffer(buffer, src_video_info, dst_video_info, sample, &total_size, &dst_buffer);
 
     if (status)
     {
@@ -1099,6 +1256,21 @@
         sample->size = 0;
         return status;
     }
+
+    if ((sgi = getenv("SteamGameId")))
+    {
+        enum fill_action action = 0;
+
+        if (!strcmp(sgi, "1449280"))
+            action |= FILL_BOTTOM;
+        else if (!strcmp(sgi, "536280"))
+            action |= FILL_RIGHT;
+
+        fill_frame_padded_bits(dst_buffer ? dst_buffer : buffer, align, dst_video_info, action);
+    }
+
+    if (dst_buffer)
+        gst_buffer_unref(dst_buffer);
 
     set_sample_flags_from_buffer(sample, buffer, total_size);
 
@@ -1214,6 +1386,13 @@
     bool discard_data;
     NTSTATUS status;
 
+    if (sample->stride != transform->attrs.output_plane_stride)
+    {
+        GST_INFO("Reconfiguring to stride %u", sample->stride);
+        transform->attrs.output_plane_stride = sample->stride;
+        push_event(transform->my_sink, gst_event_new_reconfigure());
+    }
+
     if (!transform->output_sample && !get_transform_output(transform, sample))
     {
         sample->size = 0;
@@ -1237,7 +1416,7 @@
         dst_video_info = src_video_info;
 
         /* set the desired output buffer alignment and stride on the dest video info */
-        align_video_info_planes(&transform->output_info, plane_align, &dst_video_info, &align);
+        align_video_info_planes(&transform->output_info, plane_align, sample->stride, &dst_video_info, &align);
 
         /* copy the actual output buffer alignment and stride to the src video info */
         if ((meta = gst_buffer_get_video_meta(output_buffer)))
@@ -1258,7 +1437,7 @@
 
     if (!strcmp(output_mime, "video/x-raw"))
         status = read_transform_output_video(sample, output_buffer,
-                &src_video_info, &dst_video_info);
+                &src_video_info, &dst_video_info, &align);
     else
         status = read_transform_output(sample, output_buffer);
 
