diff --git a/dlls/mfreadwrite/writer.c b/dlls/mfreadwrite/writer.c
--- a/dlls/mfreadwrite/writer.c
+++ b/dlls/mfreadwrite/writer.c
@@ -58,14 +58,14 @@
 struct stream
 {
     IMFStreamSink *stream_sink;
-    IMFTransform *converter, *encoder;
+    IMFTransform *encoder;
     MF_SINK_WRITER_STATISTICS stats;
-    struct list queue; /* struct pending_item. */
+    struct list queue;
 };
 
 struct sink_writer
 {
-    IMFSinkWriterEx IMFSinkWriterEx_iface;
+    IMFSinkWriter IMFSinkWriter_iface;
     IMFAsyncCallback events_callback;
     LONG refcount;
 
@@ -83,15 +83,14 @@
     HRESULT status;
     MF_SINK_WRITER_STATISTICS stats;
 
-    IMFAttributes *attributes;
     IMFSinkWriterCallback *callback;
 
     CRITICAL_SECTION cs;
 };
 
-static struct sink_writer *impl_from_IMFSinkWriterEx(IMFSinkWriterEx *iface)
-{
-    return CONTAINING_RECORD(iface, struct sink_writer, IMFSinkWriterEx_iface);
+static struct sink_writer *impl_from_IMFSinkWriter(IMFSinkWriter *iface)
+{
+    return CONTAINING_RECORD(iface, struct sink_writer, IMFSinkWriter_iface);
 }
 
 static struct sink_writer *impl_from_events_callback_IMFAsyncCallback(IMFAsyncCallback *iface)
@@ -166,36 +165,6 @@
 
     return S_OK;
 }
-static void stream_release_transforms(struct stream *stream)
-{
-    if (stream->converter)
-        IMFTransform_Release(stream->converter);
-    stream->converter = NULL;
-
-    if (stream->encoder)
-        IMFTransform_Release(stream->encoder);
-    stream->encoder = NULL;
-}
-
-static HRESULT stream_get_type(struct stream *stream, IMFMediaType **out_type)
-{
-    IMFMediaTypeHandler *type_handler;
-    HRESULT hr;
-
-    if (FAILED(hr = IMFStreamSink_GetMediaTypeHandler(stream->stream_sink, &type_handler)))
-        return hr;
-    hr = IMFMediaTypeHandler_GetCurrentMediaType(type_handler, out_type);
-
-    IMFMediaTypeHandler_Release(type_handler);
-    return hr;
-}
-
-static struct stream *sink_writer_get_stream(const struct sink_writer *writer, DWORD index)
-{
-    if (index >= writer->streams.count)
-        return NULL;
-    return &writer->streams.items[index];
-}
 
 static void sink_writer_release_pending_item(struct pending_item *item)
 {
@@ -205,16 +174,15 @@
     free(item);
 }
 
-static HRESULT WINAPI sink_writer_QueryInterface(IMFSinkWriterEx *iface, REFIID riid, void **out)
+static HRESULT WINAPI sink_writer_QueryInterface(IMFSinkWriter *iface, REFIID riid, void **out)
 {
     TRACE("%p, %s, %p.\n", iface, debugstr_guid(riid), out);
 
-    if (IsEqualIID(riid, &IID_IMFSinkWriterEx)
-            || IsEqualIID(riid, &IID_IMFSinkWriter)
-            || IsEqualIID(riid, &IID_IUnknown))
+    if (IsEqualIID(riid, &IID_IMFSinkWriter) ||
+            IsEqualIID(riid, &IID_IUnknown))
     {
         *out = iface;
-        IMFSinkWriterEx_AddRef(iface);
+        IMFSinkWriter_AddRef(iface);
         return S_OK;
     }
 
@@ -223,9 +191,9 @@
     return E_NOINTERFACE;
 }
 
-static ULONG WINAPI sink_writer_AddRef(IMFSinkWriterEx *iface)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static ULONG WINAPI sink_writer_AddRef(IMFSinkWriter *iface)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     ULONG refcount = InterlockedIncrement(&writer->refcount);
 
     TRACE("%p, %lu.\n", iface, refcount);
@@ -243,9 +211,9 @@
     }
 }
 
-static ULONG WINAPI sink_writer_Release(IMFSinkWriterEx *iface)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static ULONG WINAPI sink_writer_Release(IMFSinkWriter *iface)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     ULONG refcount = InterlockedDecrement(&writer->refcount);
     unsigned int i;
 
@@ -257,8 +225,6 @@
             IMFPresentationClock_Release(writer->clock);
         if (writer->sink)
             IMFMediaSink_Release(writer->sink);
-        if (writer->attributes)
-            IMFAttributes_Release(writer->attributes);
         if (writer->callback)
             IMFSinkWriterCallback_Release(writer->callback);
         for (i = 0; i < writer->streams.count; ++i)
@@ -267,7 +233,8 @@
 
             if (stream->stream_sink)
                 IMFStreamSink_Release(stream->stream_sink);
-            stream_release_transforms(stream);
+            if (stream->encoder)
+                IMFTransform_Release(stream->encoder);
             sink_writer_drop_pending_items(stream);
         }
         DeleteCriticalSection(&writer->cs);
@@ -305,9 +272,9 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_AddStream(IMFSinkWriterEx *iface, IMFMediaType *media_type, DWORD *index)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static HRESULT WINAPI sink_writer_AddStream(IMFSinkWriter *iface, IMFMediaType *media_type, DWORD *index)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     HRESULT hr = MF_E_INVALIDREQUEST;
     IMFStreamSink *stream_sink;
     DWORD id;
@@ -337,51 +304,12 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_SetInputMediaType(IMFSinkWriterEx *iface, DWORD index, IMFMediaType *type,
+static HRESULT WINAPI sink_writer_SetInputMediaType(IMFSinkWriter *iface, DWORD index, IMFMediaType *type,
         IMFAttributes *parameters)
 {
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
-    IMFMediaType *stream_type = NULL;
-    struct stream *stream;
-    DWORD flags;
-    HRESULT hr;
-
-    TRACE("%p, %lu, %p, %p.\n", iface, index, type, parameters);
-
-    if (!type)
-        return E_INVALIDARG;
-
-    EnterCriticalSection(&writer->cs);
-
-    if (!(stream = sink_writer_get_stream(writer, index)))
-    {
-        hr = MF_E_INVALIDSTREAMNUMBER;
-        goto done;
-    }
-
-    if (FAILED(hr = stream_get_type(stream, &stream_type))
-            || FAILED(hr = IMFMediaType_IsEqual(type, stream_type, &flags)))
-        goto done;
-
-    if (!(flags & MF_MEDIATYPE_EQUAL_MAJOR_TYPES))
-    {
-        hr = MF_E_INVALIDMEDIATYPE;
-        goto done;
-    }
-
-    /* Types are not compatible, create transforms. */
-    if (!(flags & MF_MEDIATYPE_EQUAL_FORMAT_DATA))
-    {
-        /* TODO: Try only using converter first, then try again with encoder. */
-        FIXME("Not implemented.\n");
-        hr = E_NOTIMPL;
-    }
-
-done:
-    if (stream_type)
-        IMFMediaType_Release(stream_type);
-    LeaveCriticalSection(&writer->cs);
-    return hr;
+    FIXME("%p, %lu, %p, %p.\n", iface, index, type, parameters);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT sink_writer_set_presentation_clock(struct sink_writer *writer)
@@ -403,9 +331,9 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_BeginWriting(IMFSinkWriterEx *iface)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static HRESULT WINAPI sink_writer_BeginWriting(IMFSinkWriter *iface)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     HRESULT hr = S_OK;
     unsigned int i;
 
@@ -429,8 +357,6 @@
                 WARN("Failed to subscribe to events for steam %u, hr %#lx.\n", i, hr);
             }
 
-            if (stream->converter)
-                IMFTransform_ProcessMessage(stream->converter, MFT_MESSAGE_NOTIFY_BEGIN_STREAMING, 0);
             if (stream->encoder)
                 IMFTransform_ProcessMessage(stream->encoder, MFT_MESSAGE_NOTIFY_BEGIN_STREAMING, 0);
         }
@@ -444,6 +370,12 @@
     LeaveCriticalSection(&writer->cs);
 
     return hr;
+}
+
+static struct stream * sink_writer_get_stream(const struct sink_writer *writer, DWORD index)
+{
+    if (index >= writer->streams.count) return NULL;
+    return &writer->streams.items[index];
 }
 
 static HRESULT sink_writer_get_buffer_length(IMFSample *sample, LONGLONG *timestamp, DWORD *length)
@@ -534,8 +466,6 @@
 
     IMFStreamSink_Flush(stream->stream_sink);
 
-    if (stream->converter)
-        IMFTransform_ProcessMessage(stream->converter, MFT_MESSAGE_COMMAND_FLUSH, 0);
     if (stream->encoder)
         IMFTransform_ProcessMessage(stream->encoder, MFT_MESSAGE_COMMAND_FLUSH, 0);
 
@@ -623,9 +553,9 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_WriteSample(IMFSinkWriterEx *iface, DWORD index, IMFSample *sample)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static HRESULT WINAPI sink_writer_WriteSample(IMFSinkWriter *iface, DWORD index, IMFSample *sample)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     struct stream *stream;
     LONGLONG timestamp;
     DWORD length;
@@ -662,9 +592,9 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_SendStreamTick(IMFSinkWriterEx *iface, DWORD index, LONGLONG timestamp)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static HRESULT WINAPI sink_writer_SendStreamTick(IMFSinkWriter *iface, DWORD index, LONGLONG timestamp)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     HRESULT hr;
 
     TRACE("%p, %lu, %s.\n", iface, index, wine_dbgstr_longlong(timestamp));
@@ -681,9 +611,9 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_PlaceMarker(IMFSinkWriterEx *iface, DWORD index, void *user_context)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static HRESULT WINAPI sink_writer_PlaceMarker(IMFSinkWriter *iface, DWORD index, void *user_context)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     struct stream *stream;
     HRESULT hr;
 
@@ -705,9 +635,9 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_NotifyEndOfSegment(IMFSinkWriterEx *iface, DWORD index)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static HRESULT WINAPI sink_writer_NotifyEndOfSegment(IMFSinkWriter *iface, DWORD index)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     HRESULT hr = S_OK;
     unsigned int i;
 
@@ -736,9 +666,9 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_Flush(IMFSinkWriterEx *iface, DWORD index)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static HRESULT WINAPI sink_writer_Flush(IMFSinkWriter *iface, DWORD index)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     HRESULT hr = S_OK;
     unsigned int i;
 
@@ -767,7 +697,7 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_Finalize(IMFSinkWriterEx *iface)
+static HRESULT WINAPI sink_writer_Finalize(IMFSinkWriter *iface)
 {
     FIXME("%p.\n", iface);
 
@@ -793,10 +723,10 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_GetServiceForStream(IMFSinkWriterEx *iface, DWORD index, REFGUID service,
+static HRESULT WINAPI sink_writer_GetServiceForStream(IMFSinkWriter *iface, DWORD index, REFGUID service,
         REFIID riid, void **object)
 {
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     HRESULT hr = E_UNEXPECTED;
     struct stream *stream;
 
@@ -810,9 +740,6 @@
     {
         if (stream->encoder)
             hr = sink_writer_get_service(stream->encoder, service, riid, object);
-        else if (stream->converter)
-            hr = sink_writer_get_service(stream->converter, service, riid, object);
-
         if (FAILED(hr))
             hr = sink_writer_get_service(stream->stream_sink, service, riid, object);
     }
@@ -824,9 +751,9 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_GetStatistics(IMFSinkWriterEx *iface, DWORD index, MF_SINK_WRITER_STATISTICS *stats)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
+static HRESULT WINAPI sink_writer_GetStatistics(IMFSinkWriter *iface, DWORD index, MF_SINK_WRITER_STATISTICS *stats)
+{
+    struct sink_writer *writer = impl_from_IMFSinkWriter(iface);
     struct stream *stream;
     HRESULT hr = S_OK;
 
@@ -854,55 +781,7 @@
     return hr;
 }
 
-static HRESULT WINAPI sink_writer_GetTransformForStream(IMFSinkWriterEx *iface,
-        DWORD stream_index, DWORD transform_index, GUID *out_category, IMFTransform **out_transform)
-{
-    struct sink_writer *writer = impl_from_IMFSinkWriterEx(iface);
-    IMFTransform *transform = NULL;
-    struct stream *stream;
-    GUID category = {};
-    HRESULT hr;
-
-    TRACE("%p, %#lx, %#lx, %p, %p.\n", iface, stream_index, transform_index, out_category, out_transform);
-
-    if (!out_transform)
-        return E_POINTER;
-    *out_transform = NULL;
-
-    EnterCriticalSection(&writer->cs);
-
-    if (!(stream = sink_writer_get_stream(writer, stream_index)))
-    {
-        LeaveCriticalSection(&writer->cs);
-        return MF_E_INVALIDSTREAMNUMBER;
-    }
-
-    if (stream->converter && transform_index == 0)
-    {
-        transform = stream->converter;
-        category = MFT_CATEGORY_VIDEO_PROCESSOR;
-    }
-    else if ((stream->converter && transform_index == 1)
-            || (!stream->converter && transform_index == 0))
-    {
-        transform = stream->encoder;
-        category = MFT_CATEGORY_VIDEO_ENCODER;
-    }
-
-    hr = MF_E_INVALIDINDEX;
-    if (transform)
-    {
-        IMFTransform_AddRef((*out_transform = transform));
-        if (out_category)
-            *out_category = category;
-        hr = S_OK;
-    }
-
-    LeaveCriticalSection(&writer->cs);
-    return hr;
-}
-
-static const IMFSinkWriterExVtbl sink_writer_vtbl =
+static const IMFSinkWriterVtbl sink_writer_vtbl =
 {
     sink_writer_QueryInterface,
     sink_writer_AddRef,
@@ -918,7 +797,6 @@
     sink_writer_Finalize,
     sink_writer_GetServiceForStream,
     sink_writer_GetStatistics,
-    sink_writer_GetTransformForStream,
 };
 
 static HRESULT WINAPI sink_writer_callback_QueryInterface(IMFAsyncCallback *iface,
@@ -940,13 +818,13 @@
 static ULONG WINAPI sink_writer_events_callback_AddRef(IMFAsyncCallback *iface)
 {
     struct sink_writer *writer = impl_from_events_callback_IMFAsyncCallback(iface);
-    return IMFSinkWriterEx_AddRef(&writer->IMFSinkWriterEx_iface);
+    return IMFSinkWriter_AddRef(&writer->IMFSinkWriter_iface);
 }
 
 static ULONG WINAPI sink_writer_events_callback_Release(IMFAsyncCallback *iface)
 {
     struct sink_writer *writer = impl_from_events_callback_IMFAsyncCallback(iface);
-    return IMFSinkWriterEx_Release(&writer->IMFSinkWriterEx_iface);
+    return IMFSinkWriter_Release(&writer->IMFSinkWriter_iface);
 }
 
 static HRESULT WINAPI sink_writer_callback_GetParameters(IMFAsyncCallback *iface,
@@ -1090,7 +968,7 @@
     if (!(object = calloc(1, sizeof(*object))))
         return E_OUTOFMEMORY;
 
-    object->IMFSinkWriterEx_iface.lpVtbl = &sink_writer_vtbl;
+    object->IMFSinkWriter_iface.lpVtbl = &sink_writer_vtbl;
     object->events_callback.lpVtbl = &sink_writer_events_callback_vtbl;
     object->refcount = 1;
     object->sink = sink;
@@ -1100,19 +978,18 @@
 
     if (attributes)
     {
-        IMFAttributes_AddRef((object->attributes = attributes));
         IMFAttributes_GetUnknown(attributes, &MF_SINK_WRITER_ASYNC_CALLBACK,
                 &IID_IMFSinkWriterCallback, (void **)&object->callback);
     }
 
     if (FAILED(hr = sink_writer_initialize_existing_streams(object, sink)))
     {
-        IMFSinkWriterEx_Release(&object->IMFSinkWriterEx_iface);
+        IMFSinkWriter_Release(&object->IMFSinkWriter_iface);
         return hr;
     }
 
-    hr = IMFSinkWriterEx_QueryInterface(&object->IMFSinkWriterEx_iface, riid, out);
-    IMFSinkWriterEx_Release(&object->IMFSinkWriterEx_iface);
+    hr = IMFSinkWriter_QueryInterface(&object->IMFSinkWriter_iface, riid, out);
+    IMFSinkWriter_Release(&object->IMFSinkWriter_iface);
     return hr;
 }
 
