diff --git a/dlls/ntdll/unix/signal_arm.c b/dlls/ntdll/unix/signal_arm.c
--- a/dlls/ntdll/unix/signal_arm.c
+++ b/dlls/ntdll/unix/signal_arm.c
@@ -229,17 +229,46 @@
     UINT                  fpscr;          /* 048 */
     struct syscall_frame *prev_frame;     /* 04c */
     void                 *syscall_cfa;    /* 050 */
-    UINT                  syscall_id;     /* 054 */
-    UINT                  align[2];       /* 058 */
+    UINT                  align[3];       /* 054 */
     ULONGLONG             d[32];          /* 060 */
 };
 
 C_ASSERT( sizeof( struct syscall_frame ) == 0x160);
 
+struct arm_thread_data
+{
+    SYSTEM_SERVICE_TABLE *syscall_table; /* 1d4 syscall table */
+    struct syscall_frame *syscall_frame; /* 1d8 frame pointer on syscall entry */
+};
+
+C_ASSERT( sizeof(struct arm_thread_data) <= sizeof(((struct ntdll_thread_data *)0)->cpu_data) );
+C_ASSERT( offsetof( TEB, GdiTebBatch ) + offsetof( struct arm_thread_data, syscall_table ) == 0x1d4 );
+C_ASSERT( offsetof( TEB, GdiTebBatch ) + offsetof( struct arm_thread_data, syscall_frame ) == 0x1d8 );
+
+static inline struct arm_thread_data *arm_thread_data(void)
+{
+    return (struct arm_thread_data *)ntdll_get_thread_data()->cpu_data;
+}
+
+static BOOL is_inside_syscall( ucontext_t *sigcontext )
+{
+    return ((char *)SP_sig(sigcontext) >= (char *)ntdll_get_thread_data()->kernel_stack &&
+            (char *)SP_sig(sigcontext) <= (char *)arm_thread_data()->syscall_frame);
+}
+
 
 void set_process_instrumentation_callback( void *callback )
 {
     if (callback) FIXME( "Not supported.\n" );
+}
+
+
+/***********************************************************************
+ *           unwind_builtin_dll
+ */
+NTSTATUS unwind_builtin_dll( void *args )
+{
+    return STATUS_UNSUCCESSFUL;
 }
 
 
@@ -370,7 +399,7 @@
     NTSTATUS status = NtSetContextThread( GetCurrentThread(), context );
 
     if (!status && (context->ContextFlags & CONTEXT_INTEGER) == CONTEXT_INTEGER)
-        get_syscall_frame()->restore_flags |= CONTEXT_INTEGER;
+        arm_thread_data()->syscall_frame->restore_flags |= CONTEXT_INTEGER;
     return status;
 }
 
@@ -380,7 +409,7 @@
  */
 void *get_native_context( CONTEXT *context )
 {
-    return is_old_wow64() ? NULL : context;
+    return context;
 }
 
 
@@ -389,7 +418,7 @@
  */
 void *get_wow_context( CONTEXT *context )
 {
-    return is_old_wow64() ? context : NULL;
+    return NULL;
 }
 
 
@@ -400,7 +429,7 @@
 NTSTATUS WINAPI NtSetContextThread( HANDLE handle, const CONTEXT *context )
 {
     NTSTATUS ret;
-    struct syscall_frame *frame = get_syscall_frame();
+    struct syscall_frame *frame = arm_thread_data()->syscall_frame;
     DWORD flags = context->ContextFlags & ~CONTEXT_ARM;
     BOOL self = (handle == GetCurrentThread());
 
@@ -449,7 +478,7 @@
  */
 NTSTATUS WINAPI NtGetContextThread( HANDLE handle, CONTEXT *context )
 {
-    struct syscall_frame *frame = get_syscall_frame();
+    struct syscall_frame *frame = arm_thread_data()->syscall_frame;
     DWORD needed_flags = context->ContextFlags & ~CONTEXT_ARM;
     BOOL self = (handle == GetCurrentThread());
 
@@ -562,14 +591,12 @@
 /***********************************************************************
  *           call_user_apc_dispatcher
  */
-NTSTATUS call_user_apc_dispatcher( CONTEXT *context, unsigned int flags, ULONG_PTR arg1, ULONG_PTR arg2, ULONG_PTR arg3,
+NTSTATUS call_user_apc_dispatcher( CONTEXT *context, ULONG_PTR arg1, ULONG_PTR arg2, ULONG_PTR arg3,
                                    PNTAPCFUNC func, NTSTATUS status )
 {
-    struct syscall_frame *frame = get_syscall_frame();
+    struct syscall_frame *frame = arm_thread_data()->syscall_frame;
     ULONG sp = context ? context->Sp : frame->sp;
     struct apc_stack_layout *stack;
-
-    if (flags) FIXME( "flags %#x are not supported.\n", flags );
 
     sp &= ~7;
     stack = (struct apc_stack_layout *)sp - 1;
@@ -602,7 +629,7 @@
  */
 void call_raise_user_exception_dispatcher(void)
 {
-    struct syscall_frame *frame = get_syscall_frame();
+    struct syscall_frame *frame = arm_thread_data()->syscall_frame;
 
     frame->sp += 16;
     frame->pc = (DWORD)pKiRaiseUserExceptionDispatcher;
@@ -615,7 +642,7 @@
 NTSTATUS call_user_exception_dispatcher( EXCEPTION_RECORD *rec, CONTEXT *context )
 {
     struct exc_stack_layout *stack;
-    struct syscall_frame *frame = get_syscall_frame();
+    struct syscall_frame *frame = arm_thread_data()->syscall_frame;
     NTSTATUS status = NtSetContextThread( GetCurrentThread(), context );
 
     if (status) return status;
@@ -657,24 +684,13 @@
                    "vstm r5, {d8-d15}\n\t"
                    "str r6, [r5, #0x80]\n\t"
                    "sub sp, sp, #0x160\n\t"   /* sizeof(struct syscall_frame) + registers */
-                   "ldr ip, [r4, #0x218]\n\t" /* thread_data->syscall_frame */
-                   "str ip, [sp, #0x4c]\n\t"  /* frame->prev_frame */
+                   "ldr r5, [r4, #0x1d8]\n\t" /* arm_thread_data()->syscall_frame */
+                   "str r5, [sp, #0x4c]\n\t"  /* frame->prev_frame */
                    "str r7, [sp, #0x50]\n\t"  /* frame->syscall_cfa */
-                   "str sp, [r4, #0x218]\n\t" /* thread_data->syscall_frame */
-                   "ldr r7, [r4, #0x21c]\n\t" /* thread_data->syscall_trace */
-                   "cbnz r7, 1f\n\t"
+                   "str sp, [r4, #0x1d8]\n\t" /* arm_thread_data()->syscall_frame */
                    /* switch to user stack */
                    "mov sp, r0\n\t"
-                   "bx r3\n"
-                   "1:\tmov r6, r0\n\t"       /* user_sp */
-                   "ldr r1, [r6]\n\t"         /* args */
-                   "ldr r2, [r6, #4]\n\t"     /* len */
-                   "ldr r0, [r6, #8]\n\t"     /* id */
-                   "str r0, [r5, #0x84]\n\t"
-                   "mov r4, r3\n\t"           /* func */
-                   "bl " __ASM_NAME("trace_usercall") "\n\t"
-                   "mov sp, r6\n\t"
-                   "bx r4" )
+                   "bx r3" )
 
 
 /***********************************************************************
@@ -683,20 +699,14 @@
 extern void DECLSPEC_NORETURN user_mode_callback_return( void *ret_ptr, ULONG ret_len,
                                                          NTSTATUS status, TEB *teb );
 __ASM_GLOBAL_FUNC( user_mode_callback_return,
-                   "ldr r4, [r3, #0x218]\n\t" /* thread_data->syscall_frame */
+                   "ldr r4, [r3, #0x1d8]\n\t" /* arm_thread_data()->syscall_frame */
                    "ldr r5, [r4, #0x4c]\n\t"  /* frame->prev_frame */
-                   "str r5, [r3, #0x218]\n\t" /* thread_data->syscall_frame */
+                   "str r5, [r3, #0x1d8]\n\t" /* arm_thread_data()->syscall_frame */
                    "add r5, r4, #0x160\n\t"
                    "vldm r5, {d8-d15}\n\t"
                    "ldr r6, [r5, #0x80]\n\t"
                    "vmsr fpscr, r6\n\t"
-                   "ldr r7, [r3, #0x21c]\n\t" /* thread_data->syscall_trace */
-                   "cbz r7, 1f\n\t"
-                   "stm r5, {r0-r3}\n\t"
-                   "ldr r3, [r5, #0x84]\n\t"  /* id */
-                   "bl " __ASM_NAME("trace_userret") "\n\t"
-                   "ldm r5, {r0-r3}\n"
-                   "1:\tadd r5, r5, #0x90\n\t"
+                   "add r5, r5, #0x90\n\t"
                    "mov sp, r5\n\t"
                    "pop {r4-r7}\n\t"          /* ret_ptr, ret_len, teb, exception_list */
                    "str r7, [r3]\n\t"         /* teb->Tib.ExceptionList */
@@ -735,7 +745,7 @@
  */
 NTSTATUS KeUserModeCallback( ULONG id, const void *args, ULONG len, void **ret_ptr, ULONG *ret_len )
 {
-    struct syscall_frame *frame = get_syscall_frame();
+    struct syscall_frame *frame = arm_thread_data()->syscall_frame;
     ULONG sp = (frame->sp - offsetof( struct callback_stack_layout, args_data[len] ) - 8) & ~7;
     struct callback_stack_layout *stack = (struct callback_stack_layout *)sp;
 
@@ -758,7 +768,7 @@
  */
 NTSTATUS WINAPI NtCallbackReturn( void *ret_ptr, ULONG ret_len, NTSTATUS status )
 {
-    if (!get_syscall_frame()->prev_frame) return STATUS_NO_CALLBACK_ACTIVE;
+    if (!arm_thread_data()->syscall_frame->prev_frame) return STATUS_NO_CALLBACK_ACTIVE;
     user_mode_callback_return( ret_ptr, ret_len, status, NtCurrentTeb() );
 }
 
@@ -770,25 +780,29 @@
  */
 static BOOL handle_syscall_fault( ucontext_t *context, EXCEPTION_RECORD *rec )
 {
-    struct syscall_frame *frame = get_syscall_frame();
+    struct syscall_frame *frame = arm_thread_data()->syscall_frame;
     UINT i;
 
-    if (!is_inside_syscall( SP_sig(context) )) return FALSE;
-
-    TRACE( "code=%x flags=%x addr=%p pc=%08x\n",
+    if (!is_inside_syscall( context )) return FALSE;
+
+    TRACE( "code=%lx flags=%lx addr=%p pc=%08lx\n",
            rec->ExceptionCode, rec->ExceptionFlags, rec->ExceptionAddress, (DWORD)PC_sig(context) );
     for (i = 0; i < rec->NumberParameters; i++)
         TRACE( " info[%d]=%08lx\n", i, rec->ExceptionInformation[i] );
 
-    TRACE( " r0=%08x r1=%08x r2=%08x r3=%08x r4=%08x r5=%08x\n",
+    TRACE( " r0=%08lx r1=%08lx r2=%08lx r3=%08lx r4=%08lx r5=%08lx\n",
            (DWORD)REGn_sig(0, context), (DWORD)REGn_sig(1, context), (DWORD)REGn_sig(2, context),
            (DWORD)REGn_sig(3, context), (DWORD)REGn_sig(4, context), (DWORD)REGn_sig(5, context) );
-    TRACE( " r6=%08x r7=%08x r8=%08x r9=%08x r10=%08x r11=%08x\n",
+    TRACE( " r6=%08lx r7=%08lx r8=%08lx r9=%08lx r10=%08lx r11=%08lx\n",
            (DWORD)REGn_sig(6, context), (DWORD)REGn_sig(7, context), (DWORD)REGn_sig(8, context),
            (DWORD)REGn_sig(9, context), (DWORD)REGn_sig(10, context), (DWORD)FP_sig(context) );
-    TRACE( " r12=%08x sp=%08x lr=%08x pc=%08x cpsr=%08x\n",
+    TRACE( " r12=%08lx sp=%08lx lr=%08lx pc=%08lx cpsr=%08lx\n",
            (DWORD)IP_sig(context), (DWORD)SP_sig(context), (DWORD)LR_sig(context),
            (DWORD)PC_sig(context), (DWORD)CPSR_sig(context) );
+
+    if (rec->ExceptionCode == STATUS_ACCESS_VIOLATION
+            && is_inside_syscall_stack_guard( (char *)rec->ExceptionInformation[1] ))
+        ERR_(seh)( "Syscall stack overrun.\n ");
 
     if (ntdll_get_thread_data()->jmp_buf)
     {
@@ -800,14 +814,11 @@
     }
     else
     {
-        TRACE( "returning to user mode ip=%08x ret=%08x\n", frame->pc, rec->ExceptionCode );
-        REGn_sig(0, context) = rec->ExceptionCode;
-        REGn_sig(1, context) = (DWORD)NtCurrentTeb();
-        REGn_sig(8, context) = (DWORD)frame;
+        TRACE( "returning to user mode ip=%08x ret=%08lx\n", frame->pc, rec->ExceptionCode );
+        REGn_sig(0, context) = (DWORD)frame;
+        REGn_sig(1, context) = rec->ExceptionCode;
         PC_sig(context)      = (DWORD)__wine_syscall_dispatcher_return;
     }
-    if (PC_sig(context) & 1) CPSR_sig(context) |= 0x20;
-    else CPSR_sig(context) &= ~0x20;
     return TRUE;
 }
 
@@ -1000,9 +1011,7 @@
  */
 static void quit_handler( int signal, siginfo_t *siginfo, void *sigcontext )
 {
-    ucontext_t *ucontext = sigcontext;
-
-    if (!is_inside_syscall( SP_sig(ucontext) )) user_mode_abort_thread( 0, get_syscall_frame() );
+    if (!is_inside_syscall( sigcontext )) user_mode_abort_thread( 0, arm_thread_data()->syscall_frame );
     abort_thread(0);
 }
 
@@ -1014,10 +1023,9 @@
  */
 static void usr1_handler( int signal, siginfo_t *siginfo, void *sigcontext )
 {
-    ucontext_t *ucontext = sigcontext;
     CONTEXT context;
 
-    if (is_inside_syscall( SP_sig(ucontext) ))
+    if (is_inside_syscall( sigcontext ))
     {
         context.ContextFlags = CONTEXT_FULL | CONTEXT_EXCEPTION_REQUEST;
         NtGetContextThread( GetCurrentThread(), &context );
@@ -1037,7 +1045,17 @@
 /**********************************************************************
  *           get_thread_ldt_entry
  */
-NTSTATUS get_thread_ldt_entry( HANDLE handle, THREAD_DESCRIPTOR_INFORMATION *info, ULONG len )
+NTSTATUS get_thread_ldt_entry( HANDLE handle, void *data, ULONG len, ULONG *ret_len )
+{
+    return STATUS_NOT_IMPLEMENTED;
+}
+
+
+/******************************************************************************
+ *           NtSetLdtEntries   (NTDLL.@)
+ *           ZwSetLdtEntries   (NTDLL.@)
+ */
+NTSTATUS WINAPI NtSetLdtEntries( ULONG sel1, LDT_ENTRY entry1, ULONG sel2, LDT_ENTRY entry2 )
 {
     return STATUS_NOT_IMPLEMENTED;
 }
@@ -1075,12 +1093,9 @@
 void signal_init_process(void)
 {
     struct sigaction sig_act;
-    struct ntdll_thread_data *thread_data = ntdll_get_thread_data();
-    void *kernel_stack = (char *)thread_data->kernel_stack + kernel_stack_size;
-
-    thread_data->syscall_frame = (struct syscall_frame *)kernel_stack - 1;
-
-    signal_alloc_thread( NtCurrentTeb() );
+    void *kernel_stack = (char *)ntdll_get_thread_data()->kernel_stack + kernel_stack_size;
+
+    arm_thread_data()->syscall_frame = (struct syscall_frame *)kernel_stack - 1;
 
     sig_act.sa_mask = server_block_set;
     sig_act.sa_flags = SA_RESTART | SA_SIGINFO | SA_ONSTACK;
@@ -1110,12 +1125,15 @@
 
 
 /***********************************************************************
- *           init_syscall_frame
- */
-void init_syscall_frame( LPTHREAD_START_ROUTINE entry, void *arg, BOOL suspend, TEB *teb )
-{
-    struct syscall_frame *frame = ((struct ntdll_thread_data *)&teb->GdiTebBatch)->syscall_frame;
+ *           call_init_thunk
+ */
+void call_init_thunk( LPTHREAD_START_ROUTINE entry, void *arg, BOOL suspend, TEB *teb,
+                      struct syscall_frame *frame, void *syscall_cfa )
+{
+    struct arm_thread_data *thread_data = (struct arm_thread_data *)&teb->GdiTebBatch;
     CONTEXT *ctx, context = { CONTEXT_ALL };
+
+    thread_data->syscall_table = KeServiceDescriptorTable;
 
     context.R0 = (DWORD)entry;
     context.R1 = (DWORD)arg;
@@ -1133,13 +1151,17 @@
     ctx = (CONTEXT *)((ULONG_PTR)context.Sp & ~15) - 1;
     *ctx = context;
     ctx->ContextFlags = CONTEXT_FULL;
-    signal_set_full_context( ctx );
+    memset( frame, 0, sizeof(*frame) );
+    NtSetContextThread( GetCurrentThread(), ctx );
 
     frame->sp = (DWORD)ctx;
     frame->pc = (DWORD)pLdrInitializeThunk;
     frame->r0 = (DWORD)ctx;
+    frame->restore_flags |= CONTEXT_INTEGER;
+    frame->syscall_cfa    = syscall_cfa;
 
     pthread_sigmask( SIG_UNBLOCK, &server_block_set, NULL );
+    __wine_syscall_dispatcher_return( frame, 0 );
 }
 
 
@@ -1152,19 +1174,14 @@
                    "add r7, sp, #0x28\n\t"    /* syscall_cfa */
                    "mcr p15, 0, r3, c13, c0, 2\n\t" /* set teb register */
                    /* set syscall frame */
-                   "ldr r6, [r3, #0x218]\n\t" /* thread_data->syscall_frame */
+                   "ldr r6, [r3, #0x1d8]\n\t" /* arm_thread_data()->syscall_frame */
                    "cbnz r6, 1f\n\t"
                    "sub r6, sp, #0x160\n\t"   /* sizeof(struct syscall_frame) */
-                   "str r6, [r3, #0x218]\n\t" /* thread_data->syscall_frame */
-                   "1:\tmov r4, #0\n\t"
-                   "str r4, [r6, #0x44]\n\t"  /* frame->restore_flags */
-                   "str r4, [r6, #0x4c]\n\t"  /* frame->prev_frame */
-                   "str r7, [r6, #0x50]\n\t"  /* frame->syscall_cfa */
+                   "str r6, [r3, #0x1d8]\n\t" /* arm_thread_data()->syscall_frame */
                    /* switch to kernel stack */
-                   "mov sp, r6\n\t"
-                   "mov r8, r6\n\t"
-                   "bl init_syscall_frame\n\t"
-                   "b " __ASM_LOCAL_LABEL("__wine_syscall_dispatcher_return") )
+                   "1:\tmov sp, r6\n\t"
+                   "push {r6,r7}\n\t"
+                   "bl " __ASM_NAME("call_init_thunk") )
 
 
 /***********************************************************************
@@ -1173,7 +1190,7 @@
 __ASM_GLOBAL_FUNC( __wine_syscall_dispatcher,
                    __ASM_EHABI(".cantunwind\n\t")
                    "mrc p15, 0, r2, c13, c0, 2\n\t" /* NtCurrentTeb() */
-                   "ldr r1, [r2, #0x218]\n\t"       /* thread_data->syscall_frame */
+                   "ldr r1, [r2, #0x1d8]\n\t"       /* arm_thread_data()->syscall_frame */
                    "add r0, r1, #0x10\n\t"
                    "stm r0, {r4-r12,lr}\n\t"
                    "str sp, [r1, #0x38]\n\t"
@@ -1185,7 +1202,6 @@
                    "str r0, [r1, #0x44]\n\t"        /* frame->restore_flags */
                    "vmrs r0, fpscr\n\t"
                    "str r0, [r1, #0x48]\n\t"
-                   "str ip, [r1, #0x54]\n\t"        /* frame->syscall_id */
                    "add r0, r1, #0x60\n\t"
                    "vstm r0, {d0-d15}\n\t"
                    "mov r6, sp\n\t"
@@ -1203,13 +1219,13 @@
                    __ASM_CFI(".cfi_offset r11,-0x0c\n\t")
                    __ASM_CFI(".cfi_offset r12,-0x08\n\t")
                    __ASM_CFI(".cfi_offset lr,-0x04\n\t")
-                   "ldr r5, [r2, #0x214]\n\t"       /* thread_data->syscall_table */
+                   "ldr r5, [r2, #0x1d4]\n\t"       /* arm_thread_data()->syscall_table */
                    "ubfx r4, ip, #12, #2\n\t"       /* syscall table number */
                    "bfc ip, #12, #20\n\t"           /* syscall number */
                    "add r4, r5, r4, lsl #4\n\t"
                    "ldr r5, [r4, #8]\n\t"           /* table->ServiceLimit */
                    "cmp ip, r5\n\t"
-                   "bcs " __ASM_LOCAL_LABEL("bad_syscall") "\n\t"
+                   "bcs 5f\n\t"
                    "ldr r5, [r4, #12]\n\t"          /* table->ArgumentTable */
                    "ldrb r5, [r5, ip]\n\t"
                    "cmp r5, #16\n\t"
@@ -1222,14 +1238,12 @@
                    "ldr r0, [r6, r5]\n\t"
                    "str r0, [sp, r5]\n\t"
                    "bgt 2b\n\t"
+                   "pop {r0-r3}\n\t"                /* first 4 args are in registers */
                    "ldr r5, [r4]\n\t"               /* table->ServiceTable */
-                   "ldr r5, [r5, ip, lsl #2]\n\t"
-                   "ldr r7, [r2, #0x21c]\n\t"       /* thread_data->syscall_trace */
-                   "cbnz r7, " __ASM_LOCAL_LABEL("trace_syscall") "\n\t"
-                   "pop {r0-r3}\n\t"                /* first 4 args are in registers */
-                   "blx r5\n"
+                   "ldr ip, [r5, ip, lsl #2]\n\t"
+                   "blx ip\n"
                    __ASM_LOCAL_LABEL("__wine_syscall_dispatcher_return") ":\n\t"
-                   "ldr ip, [r8, #0x44]\n\t"        /* frame->restore_flags */
+                   "ldr ip, [r8, #0x44]\n\t"    /* frame->restore_flags */
                    "tst ip, #4\n\t"                 /* CONTEXT_FLOATING_POINT */
                    "beq 3f\n\t"
                    "ldr r4, [r8, #0x48]\n\t"
@@ -1246,35 +1260,14 @@
                    "add r8, r8, #0x10\n\t"
                    "ldm r8, {r4-r12,pc}\n"
 
-                   __ASM_LOCAL_LABEL("trace_syscall") ":\n\t"
-                   "ldr r0, [r8, #0x54]\n\t"        /* frame->syscall_id */
-                   "mov r1, sp\n\t"                 /* args */
-                   "ldr r2, [r4, #12]\n\t"          /* table->ArgumentTable */
-                   "ubfx ip, r0, #0, #12\n\t"       /* syscall number */
-                   "ldrb r2, [r2, ip]\n\t"          /* len */
-                   "bl " __ASM_NAME("trace_syscall") "\n\t"
-                   "pop {r0-r3}\n\t"
-                   "blx r5\n"
-
-                   __ASM_LOCAL_LABEL("trace_syscall_ret") ":\n\t"
-                   "mov r6, r0\n\t"                 /* retval */
-                   "ldr r0, [r8, #0x54]\n\t"        /* frame->syscall_id */
-                   "mov r1, r6\n\t"                 /* retval */
-                   "bl " __ASM_NAME("trace_sysret") "\n\t"
-                   "mov r0, r6\n\t"                 /* status */
-                   "b " __ASM_LOCAL_LABEL("__wine_syscall_dispatcher_return") "\n"
-
-                   __ASM_LOCAL_LABEL("bad_syscall") ":\n\t"
-                   "movw r0, #0x001c\n\t"           /* STATUS_INVALID_SYSTEM_SERVICE */
+                   "5:\tmovw r0, #0x000d\n\t" /* STATUS_INVALID_PARAMETER */
                    "movt r0, #0xc000\n\t"
                    "b " __ASM_LOCAL_LABEL("__wine_syscall_dispatcher_return") )
 
 __ASM_GLOBAL_FUNC( __wine_syscall_dispatcher_return,
-                   "mov sp, r8\n\t"
-                   "ldr r3, [r1, #0x21c]\n\t"       /* thread_data->syscall_trace */
-                   "cbnz r3, 1f\n\t"
-                   "b " __ASM_LOCAL_LABEL("__wine_syscall_dispatcher_return") "\n"
-                   "1:\tb " __ASM_LOCAL_LABEL("trace_syscall_ret") )
+                   "mov r8, r0\n\t"
+                   "mov r0, r1\n\t"
+                   "b " __ASM_LOCAL_LABEL("__wine_syscall_dispatcher_return") )
 
 
 /***********************************************************************
@@ -1283,7 +1276,7 @@
 __ASM_GLOBAL_FUNC( __wine_unix_call_dispatcher,
                    __ASM_EHABI(".cantunwind\n\t")
                    "mrc p15, 0, r1, c13, c0, 2\n\t" /* NtCurrentTeb() */
-                   "ldr r1, [r1, #0x218]\n\t"       /* thread_data->syscall_frame */
+                   "ldr r1, [r1, #0x1d8]\n\t"       /* arm_thread_data()->syscall_frame */
                    "add ip, r1, #0x10\n\t"
                    "stm ip, {r4-r12,lr}\n\t"
                    "str sp, [r1, #0x38]\n\t"
