diff --git a/dlls/comctl32/toolbar.c b/dlls/comctl32/toolbar.c
--- a/dlls/comctl32/toolbar.c
+++ b/dlls/comctl32/toolbar.c
@@ -79,6 +79,8 @@
 #include "winnls.h"
 #include "commctrl.h"
 #include "comctl32.h"
+#include "uxtheme.h"
+#include "vssym32.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(toolbar);
@@ -140,6 +142,7 @@
     INT      iListGap;        /* default gap between text and image for toolbar with list style */
     HFONT    hDefaultFont;
     HFONT    hFont;           /* text font */
+    HTHEME   hTheme;          /* theme */
     HIMAGELIST himlInt;       /* image list created internally */
     PIMLENTRY *himlDef;       /* default image list array */
     INT       cimlDef;        /* default image list array count */
@@ -169,9 +172,6 @@
     TBUTTON_INFO *buttons;      /* pointer to button array */
     LPWSTR       *strings;      /* pointer to string array */
     TBITMAP_INFO *bitmaps;
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME   hTheme;            /* theme */
-#endif
 } TOOLBAR_INFO, *PTOOLBAR_INFO;
 
 
@@ -233,7 +233,8 @@
 #define GETHOTIMAGELIST(infoPtr, id) TOOLBAR_GetImageList(infoPtr->himlHot, infoPtr->cimlHot, id)
 #define GETDISIMAGELIST(infoPtr, id) TOOLBAR_GetImageList(infoPtr->himlDis, infoPtr->cimlDis, id)
 
-static LRESULT TOOLBAR_DoEraseBackground(TOOLBAR_INFO *infoPtr, WPARAM wParam, LPARAM lParam);
+static const WCHAR themeClass[] = L"Toolbar";
+
 static BOOL TOOLBAR_GetButtonInfo(const TOOLBAR_INFO *infoPtr, NMTOOLBARW *nmtb);
 static BOOL TOOLBAR_IsButtonRemovable(const TOOLBAR_INFO *infoPtr, int iItem, const CUSTOMBUTTON *btnInfo);
 static HIMAGELIST TOOLBAR_GetImageList(const PIMLENTRY *pies, INT cies, INT id);
@@ -264,15 +265,6 @@
 {
     return (TOOLBAR_HasDropDownArrows( infoPtr->dwExStyle ) && (btnPtr->fsStyle & BTNS_DROPDOWN)) ||
         (btnPtr->fsStyle & BTNS_WHOLEDROPDOWN);
-}
-
-static BOOL TOOLBAR_IsThemed(const TOOLBAR_INFO *infoPtr)
-{
-#if __WINE_COMCTL32_VERSION == 6
-    return !!infoPtr->hTheme;
-#else
-    return FALSE;
-#endif
 }
 
 static LPWSTR
@@ -782,7 +774,7 @@
     }
     else if (tbcd->nmcd.uItemState & CDIS_CHECKED ||
       ((tbcd->nmcd.uItemState & CDIS_HOT) 
-      && ((infoPtr->dwStyle & TBSTYLE_FLAT) || TOOLBAR_IsThemed(infoPtr))))
+      && ((infoPtr->dwStyle & TBSTYLE_FLAT) || infoPtr->hTheme)))
     {
         /* if hot, attempt to draw with hot image list, if fails, 
            use default image list */
@@ -815,7 +807,7 @@
 
 /* draws a blank frame for a toolbar button */
 static void
-TOOLBAR_DrawFlatButtonFrame(const TOOLBAR_INFO *infoPtr, const NMTBCUSTOMDRAW *tbcd, const RECT *rect, DWORD dwItemCDFlag)
+TOOLBAR_DrawFrame(const TOOLBAR_INFO *infoPtr, const NMTBCUSTOMDRAW *tbcd, const RECT *rect, DWORD dwItemCDFlag)
 {
     HDC hdc = tbcd->nmcd.hdc;
     RECT rc = *rect;
@@ -849,7 +841,7 @@
 }
 
 static void
-TOOLBAR_DrawFlatSepDropDownArrow(const TOOLBAR_INFO *infoPtr, const NMTBCUSTOMDRAW *tbcd, RECT *rcArrow, BOOL bDropDownPressed, DWORD dwItemCDFlag)
+TOOLBAR_DrawSepDDArrow(const TOOLBAR_INFO *infoPtr, const NMTBCUSTOMDRAW *tbcd, RECT *rcArrow, BOOL bDropDownPressed, DWORD dwItemCDFlag)
 {
     HDC hdc = tbcd->nmcd.hdc;
     int offset = 0;
@@ -886,106 +878,12 @@
         TOOLBAR_DrawArrow(hdc, rcArrow->left + offset, rcArrow->top + offset + (rcArrow->bottom - rcArrow->top - ARROW_HEIGHT) / 2, comctl32_color.clrBtnText);
 }
 
-static void
-TOOLBAR_DrawSeparator (const TOOLBAR_INFO *infoPtr, TBUTTON_INFO *btnPtr, HDC hdc, const RECT *rect)
-{
-#if __WINE_COMCTL32_VERSION == 6
-    if (infoPtr->hTheme)
-    {
-        int part = (infoPtr->dwStyle & CCS_VERT) ? TP_SEPARATORVERT : TP_SEPARATOR;
-        DrawThemeBackground (infoPtr->hTheme, hdc, part, 0, rect, NULL);
-        return;
-    }
-#endif
-
-    /* with the FLAT style, iBitmap is the width and has already been taken into consideration in
-     * calculating the width so now we need to draw the vertical separator empirical tests show that
-     * iBitmap can/will be non-zero when drawing the vertical bar... */
-    if ((infoPtr->dwStyle & TBSTYLE_FLAT) /* && (btnPtr->iBitmap == 0) */)
-    {
-        if (infoPtr->dwStyle & CCS_VERT)
-        {
-            RECT rcsep = *rect;
-            InflateRect (&rcsep, -infoPtr->szPadding.cx, -infoPtr->szPadding.cy);
-            TOOLBAR_DrawFlatHorizontalSeparator (&rcsep, hdc, infoPtr);
-        }
-        else
-        {
-            TOOLBAR_DrawFlatSeparator (rect, hdc, infoPtr);
-        }
-    }
-    else if (btnPtr->fsStyle != BTNS_SEP)
-    {
-        FIXME("Draw some kind of separator: fsStyle=%x\n", btnPtr->fsStyle);
-    }
-}
-
-static void TOOLBAR_DrawButtonFrame (const TOOLBAR_INFO *infoPtr, TBUTTON_INFO *btnPtr,
-                                     const NMTBCUSTOMDRAW *tbcd, HDC hdc, const RECT *rect,
-                                     DWORD dwItemCDFlag, BOOL drawSepDropDownArrow)
-{
-#if __WINE_COMCTL32_VERSION == 6
-    if (infoPtr->hTheme)
-    {
-        if (!(dwItemCDFlag & TBCDRF_NOBACKGROUND))
-        {
-            int partId = drawSepDropDownArrow ? TP_SPLITBUTTON : TP_BUTTON;
-            int stateId = TS_NORMAL;
-
-            if (tbcd->nmcd.uItemState & CDIS_DISABLED)
-                stateId = TS_DISABLED;
-            else if (tbcd->nmcd.uItemState & CDIS_SELECTED)
-                stateId = TS_PRESSED;
-            else if (tbcd->nmcd.uItemState & CDIS_CHECKED)
-                stateId = (tbcd->nmcd.uItemState & CDIS_HOT) ? TS_HOTCHECKED : TS_CHECKED;
-            else if ((tbcd->nmcd.uItemState & CDIS_HOT)
-                     || (drawSepDropDownArrow && btnPtr->bDropDownPressed))
-                stateId = TS_HOT;
-
-            DrawThemeBackground(infoPtr->hTheme, hdc, partId, stateId, rect, NULL);
-        }
-
-        return;
-    }
-#endif /* __WINE_COMCTL32_VERSION == 6 */
-
-    TOOLBAR_DrawFlatButtonFrame(infoPtr, tbcd, rect, dwItemCDFlag);
-}
-
-static void TOOLBAR_DrawSepDropDownArrow (const TOOLBAR_INFO *infoPtr, TBUTTON_INFO *btnPtr,
-                                          const NMTBCUSTOMDRAW *tbcd, HDC hdc, RECT *rect,
-                                          DWORD dwItemCDFlag)
-{
-#if __WINE_COMCTL32_VERSION == 6
-    if (infoPtr->hTheme)
-    {
-        int stateId = TS_NORMAL;
-
-        if (tbcd->nmcd.uItemState & CDIS_DISABLED)
-            stateId = TS_DISABLED;
-        else if (btnPtr->bDropDownPressed || (tbcd->nmcd.uItemState & CDIS_SELECTED))
-            stateId = TS_PRESSED;
-        else if (tbcd->nmcd.uItemState & CDIS_CHECKED)
-            stateId = (tbcd->nmcd.uItemState & CDIS_HOT) ? TS_HOTCHECKED : TS_CHECKED;
-        else if (tbcd->nmcd.uItemState & CDIS_HOT)
-            stateId = TS_HOT;
-
-        DrawThemeBackground(infoPtr->hTheme, hdc, TP_DROPDOWNBUTTON, stateId, rect, NULL);
-        DrawThemeBackground(infoPtr->hTheme, hdc, TP_SPLITBUTTONDROPDOWN, stateId, rect, NULL);
-        return;
-    }
-#endif
-
-    TOOLBAR_DrawFlatSepDropDownArrow(infoPtr, tbcd, rect, btnPtr->bDropDownPressed, dwItemCDFlag);
-}
-
 /* draws a complete toolbar button */
 static void
 TOOLBAR_DrawButton (const TOOLBAR_INFO *infoPtr, TBUTTON_INFO *btnPtr, HDC hdc, DWORD dwBaseCustDraw)
 {
     DWORD dwStyle = infoPtr->dwStyle;
     BOOL hasDropDownArrow = button_has_ddarrow( infoPtr, btnPtr );
-    BOOL isThemed = TOOLBAR_IsThemed(infoPtr);
     BOOL drawSepDropDownArrow = hasDropDownArrow && 
                                 (~btnPtr->fsStyle & BTNS_WHOLEDROPDOWN);
     RECT rc, rcArrow, rcBitmap, rcText;
@@ -996,15 +894,39 @@
     INT oldBkMode;
     DWORD dwItemCustDraw;
     DWORD dwItemCDFlag;
+    HTHEME theme = infoPtr->hTheme;
 
     rc = btnPtr->rect;
     rcArrow = rc;
 
     /* separator - doesn't send NM_CUSTOMDRAW */
-    if (btnPtr->fsStyle & BTNS_SEP)
-    {
-        TOOLBAR_DrawSeparator (infoPtr, btnPtr, hdc, &rc);
-        return;
+    if (btnPtr->fsStyle & BTNS_SEP) {
+        if (theme)
+        {
+            DrawThemeBackground (theme, hdc, 
+                (dwStyle & CCS_VERT) ? TP_SEPARATORVERT : TP_SEPARATOR, 0, 
+                &rc, NULL);
+        }
+        else
+        /* with the FLAT style, iBitmap is the width and has already */
+        /* been taken into consideration in calculating the width    */
+        /* so now we need to draw the vertical separator             */
+        /* empirical tests show that iBitmap can/will be non-zero    */
+        /* when drawing the vertical bar...      */
+        if ((dwStyle & TBSTYLE_FLAT) /* && (btnPtr->iBitmap == 0) */) {
+            if (dwStyle & CCS_VERT) {
+                RECT rcsep = rc;
+                InflateRect(&rcsep, -infoPtr->szPadding.cx, -infoPtr->szPadding.cy);
+                TOOLBAR_DrawFlatHorizontalSeparator (&rcsep, hdc, infoPtr);
+            }
+	    else
+		TOOLBAR_DrawFlatSeparator (&rc, hdc, infoPtr);
+	}
+	else if (btnPtr->fsStyle != BTNS_SEP) {
+	    FIXME("Draw some kind of separator: fsStyle=%x\n",
+		  btnPtr->fsStyle);
+	}
+	return;
     }
 
     /* get a pointer to the text */
@@ -1120,11 +1042,11 @@
         (btnPtr->fsState & (TBSTATE_PRESSED | TBSTATE_CHECKED)))
         OffsetRect(&rcText, 1, 1);
 
-    if (!isThemed && !(tbcd.nmcd.uItemState & CDIS_HOT) &&
+    if (!theme && !(tbcd.nmcd.uItemState & CDIS_HOT) &&
         ((tbcd.nmcd.uItemState & CDIS_CHECKED) || (tbcd.nmcd.uItemState & CDIS_INDETERMINATE)))
         TOOLBAR_DrawPattern (&rc, &tbcd);
 
-    if (((infoPtr->dwStyle & TBSTYLE_FLAT) || isThemed) && (tbcd.nmcd.uItemState & CDIS_HOT))
+    if (((infoPtr->dwStyle & TBSTYLE_FLAT) || theme) && (tbcd.nmcd.uItemState & CDIS_HOT))
     {
         if ( dwItemCDFlag & TBCDRF_HILITEHOTTRACK )
         {
@@ -1138,10 +1060,50 @@
         }
     }
 
-    TOOLBAR_DrawButtonFrame(infoPtr, btnPtr, &tbcd, hdc, &rc, dwItemCDFlag, drawSepDropDownArrow);
+    if (theme)
+    {
+        if (!(dwItemCDFlag & TBCDRF_NOBACKGROUND))
+        {
+            int partId = drawSepDropDownArrow ? TP_SPLITBUTTON : TP_BUTTON;
+            int stateId = TS_NORMAL;
+
+            if (tbcd.nmcd.uItemState & CDIS_DISABLED)
+                stateId = TS_DISABLED;
+            else if (tbcd.nmcd.uItemState & CDIS_SELECTED)
+                stateId = TS_PRESSED;
+            else if (tbcd.nmcd.uItemState & CDIS_CHECKED)
+                stateId = (tbcd.nmcd.uItemState & CDIS_HOT) ? TS_HOTCHECKED : TS_CHECKED;
+            else if ((tbcd.nmcd.uItemState & CDIS_HOT)
+                     || (drawSepDropDownArrow && btnPtr->bDropDownPressed))
+                stateId = TS_HOT;
+
+            DrawThemeBackground(theme, hdc, partId, stateId, &rc, NULL);
+        }
+    }
+    else
+        TOOLBAR_DrawFrame(infoPtr, &tbcd, &rc, dwItemCDFlag);
 
     if (drawSepDropDownArrow)
-        TOOLBAR_DrawSepDropDownArrow(infoPtr, btnPtr, &tbcd, hdc, &rcArrow, dwItemCDFlag);
+    {
+        if (theme)
+        {
+            int stateId = TS_NORMAL;
+            
+            if (tbcd.nmcd.uItemState & CDIS_DISABLED)
+                stateId = TS_DISABLED;
+            else if (btnPtr->bDropDownPressed || (tbcd.nmcd.uItemState & CDIS_SELECTED))
+                stateId = TS_PRESSED;
+            else if (tbcd.nmcd.uItemState & CDIS_CHECKED)
+                stateId = (tbcd.nmcd.uItemState & CDIS_HOT) ? TS_HOTCHECKED : TS_CHECKED;
+            else if (tbcd.nmcd.uItemState & CDIS_HOT)
+                stateId = TS_HOT;
+                
+            DrawThemeBackground (theme, hdc, TP_DROPDOWNBUTTON, stateId, &rcArrow, NULL);
+            DrawThemeBackground (theme, hdc, TP_SPLITBUTTONDROPDOWN, stateId, &rcArrow, NULL);
+        }
+        else
+            TOOLBAR_DrawSepDDArrow(infoPtr, &tbcd, &rcArrow, btnPtr->bDropDownPressed, dwItemCDFlag);
+    }
 
     oldBkMode = SetBkMode (hdc, tbcd.nStringBkMode);
     if (!(infoPtr->dwExStyle & TBSTYLE_EX_MIXEDBUTTONS) || (btnPtr->fsStyle & BTNS_SHOWTEXT))
@@ -1188,13 +1150,6 @@
     /* the app has told us not to redraw the toolbar */
     if (!infoPtr->bDoRedraw)
         return;
-
-    /* Comctl32 v6 erases the background when TBSTYLE_TRANSPARENT is present when handling WM_PAINT,
-     * not when handling WM_ERASEBKGND */
-#if __WINE_COMCTL32_VERSION == 6
-    if (infoPtr->dwStyle & TBSTYLE_TRANSPARENT)
-        TOOLBAR_DoEraseBackground(infoPtr, (WPARAM)hdc, 0);
-#endif
 
     /* if imagelist belongs to the app, it can be changed
        by the app after setting it */
@@ -3546,7 +3501,7 @@
 static LRESULT
 TOOLBAR_GetHotItem (const TOOLBAR_INFO *infoPtr)
 {
-    if (!((infoPtr->dwStyle & TBSTYLE_FLAT) || TOOLBAR_IsThemed(infoPtr)))
+    if (!((infoPtr->dwStyle & TBSTYLE_FLAT) || infoPtr->hTheme))
 	return -1;
 
     if (infoPtr->nHotItem < 0)
@@ -5162,7 +5117,14 @@
 static LRESULT
 TOOLBAR_SetVersion (TOOLBAR_INFO *infoPtr, INT iVersion)
 {
-    return COMCTL32_SetVersion(&infoPtr->iVersion, iVersion);
+    INT iOldVersion = infoPtr->iVersion;
+
+    infoPtr->iVersion = iVersion;
+
+    if (infoPtr->iVersion >= 5)
+        TOOLBAR_SetUnicodeFormat(infoPtr, TRUE);
+
+    return iOldVersion;
 }
 
 
@@ -5340,10 +5302,7 @@
 
     SystemParametersInfoW (SPI_GETICONTITLELOGFONT, 0, &logFont, 0);
     infoPtr->hFont = infoPtr->hDefaultFont = CreateFontIndirectW (&logFont);
-
-#if __WINE_COMCTL32_VERSION == 6
-    infoPtr->hTheme = OpenThemeDataForDpi (NULL, L"Toolbar", GetDpiForWindow (hwnd));
-#endif
+    infoPtr->hTheme = OpenThemeDataForDpi (NULL, themeClass, GetDpiForWindow (hwnd));
 
     TOOLBAR_CheckStyle (infoPtr);
 
@@ -5388,9 +5347,7 @@
     /* delete default font */
     DeleteObject (infoPtr->hDefaultFont);
 
-#if __WINE_COMCTL32_VERSION == 6
     CloseThemeData (infoPtr->hTheme);
-#endif
 
     /* free toolbar info data */
     SetWindowLongPtrW (infoPtr->hwndSelf, 0, 0);
@@ -5401,7 +5358,7 @@
 
 
 static LRESULT
-TOOLBAR_DoEraseBackground (TOOLBAR_INFO *infoPtr, WPARAM wParam, LPARAM lParam)
+TOOLBAR_EraseBackground (TOOLBAR_INFO *infoPtr, WPARAM wParam, LPARAM lParam)
 {
     NMTBCUSTOMDRAW tbcd;
     INT ret = FALSE;
@@ -5435,7 +5392,7 @@
     /* If the toolbar is "transparent" then pass the WM_ERASEBKGND up
      * to my parent for processing.
      */
-    if (TOOLBAR_IsThemed(infoPtr) || (infoPtr->dwStyle & TBSTYLE_TRANSPARENT)) {
+    if (infoPtr->hTheme || (infoPtr->dwStyle & TBSTYLE_TRANSPARENT)) {
 	POINT pt, ptorig;
 	HDC hdc = (HDC)wParam;
 	HWND parent;
@@ -5471,18 +5428,6 @@
     return ret;
 }
 
-static LRESULT
-TOOLBAR_EraseBackground (TOOLBAR_INFO *infoPtr, WPARAM wParam, LPARAM lParam)
-{
-    /* In comctl32 v6, erasing the background when TBSTYLE_TRANSPARENT is present is done in the
-     * WM_PAINT handler, not WM_ERASEBKGND */
-#if __WINE_COMCTL32_VERSION == 6
-    if (infoPtr->dwStyle & TBSTYLE_TRANSPARENT)
-        return TRUE;
-#endif
-
-    return TOOLBAR_DoEraseBackground(infoPtr, wParam, lParam);
-}
 
 static inline LRESULT
 TOOLBAR_GetFont (const TOOLBAR_INFO *infoPtr)
@@ -5999,7 +5944,7 @@
     if ((infoPtr->dwStyle & TBSTYLE_TOOLTIPS) && (infoPtr->hwndToolTip == NULL))
         TOOLBAR_TooltipCreateControl(infoPtr);
     
-    if ((infoPtr->dwStyle & TBSTYLE_FLAT) || TOOLBAR_IsThemed(infoPtr)) {
+    if ((infoPtr->dwStyle & TBSTYLE_FLAT) || infoPtr->hTheme) {
         /* fill in the TRACKMOUSEEVENT struct */
         trackinfo.cbSize = sizeof(TRACKMOUSEEVENT);
         trackinfo.dwFlags = TME_QUERY;
@@ -6027,7 +5972,7 @@
 
     nHit = TOOLBAR_InternalHitTest (infoPtr, &pt, &button);
 
-    if (((infoPtr->dwStyle & TBSTYLE_FLAT) || TOOLBAR_IsThemed(infoPtr)) && (!infoPtr->bAnchor || button))
+    if (((infoPtr->dwStyle & TBSTYLE_FLAT) || infoPtr->hTheme) && (!infoPtr->bAnchor || button))
         TOOLBAR_SetHotItemEx(infoPtr, button ? nHit : TOOLBAR_NOWHERE, HICF_MOUSE);
 
     if (infoPtr->nOldHit != nHit)
@@ -6115,11 +6060,7 @@
     infoPtr->dwDTFlags = (lpcs->style & TBSTYLE_LIST) ? DT_LEFT | DT_VCENTER | DT_SINGLELINE | DT_END_ELLIPSIS: DT_CENTER | DT_END_ELLIPSIS;
     infoPtr->bAnchor = FALSE; /* no anchor highlighting */
     infoPtr->bDragOutSent = FALSE;
-#if __WINE_COMCTL32_VERSION == 6
-    infoPtr->iVersion = 6;
-#else
     infoPtr->iVersion = 0;
-#endif
     infoPtr->hwndSelf = hwnd;
     infoPtr->bDoRedraw = TRUE;
     infoPtr->clrBtnHighlight = CLR_DEFAULT;
@@ -6418,16 +6359,7 @@
         FIXME("TTN_GETDISPINFOA - should not be received; please report\n");
         return 0;
 
-    case CBEN_ENDEDITW:
-        if (infoPtr->bUnicode)
-            return SendMessageW(infoPtr->hwndNotify, WM_NOTIFY, lpnmh->idFrom, (LPARAM)lpnmh);
-        return COMCTL32_forward_notify_to_ansi_window(infoPtr->hwndNotify, lpnmh, NULL, NULL);
-
-    case CBEN_ENDEDITA:
-        return SendMessageW(infoPtr->hwndNotify, WM_NOTIFY, lpnmh->idFrom, (LPARAM)lpnmh);
-
     default:
-        WARN("Should WM_NOTIFY NMHDR code 0x%x be forwarded?\n", lpnmh->code);
         return 0;
     }
 }
@@ -6605,14 +6537,10 @@
 /* update theme after a WM_THEMECHANGED message */
 static LRESULT theme_changed (TOOLBAR_INFO *infoPtr)
 {
-#if __WINE_COMCTL32_VERSION == 6
     CloseThemeData (infoPtr->hTheme);
-    infoPtr->hTheme = OpenThemeDataForDpi (NULL, L"Toolbar", GetDpiForWindow (infoPtr->hwndSelf));
+    infoPtr->hTheme = OpenThemeDataForDpi (NULL, themeClass, GetDpiForWindow (infoPtr->hwndSelf));
     InvalidateRect (infoPtr->hwndSelf, NULL, TRUE);
     return 0;
-#else
-    return DefWindowProcW(infoPtr->hwndSelf, WM_THEMECHANGED, 0, 0);
-#endif
 }
 
 
@@ -6928,11 +6856,6 @@
 
 	case WM_GETFONT:
 		return TOOLBAR_GetFont (infoPtr);
-
-        case WM_GETOBJECT:
-            if ((LONG)lParam == OBJID_QUERYCLASSNAMEIDX)
-                return 0x1000c;
-	    return DefWindowProcW (hwnd, uMsg, wParam, lParam);
 
 	case WM_KEYDOWN:
 	    return TOOLBAR_KeyDown (infoPtr, wParam, lParam);
@@ -7045,6 +6968,13 @@
     RegisterClassW (&wndClass);
 }
 
+
+VOID
+TOOLBAR_Unregister (void)
+{
+    UnregisterClassW (TOOLBARCLASSNAMEW, NULL);
+}
+
 static HIMAGELIST TOOLBAR_InsertImageList(PIMLENTRY **pies, INT *cies, HIMAGELIST himl, INT id)
 {
     HIMAGELIST himlold;
