diff --git a/dlls/gdiplus/graphics.c b/dlls/gdiplus/graphics.c
--- a/dlls/gdiplus/graphics.c
+++ b/dlls/gdiplus/graphics.c
@@ -412,73 +412,14 @@
     gdi_dc_release(graphics, dst_hdc);
 }
 
-static GpStatus get_graphics_device_bounds(GpGraphics* graphics, GpRectF* rect)
-{
-    RECT wnd_rect;
-    GpStatus stat=Ok;
-    GpUnit unit;
-
-    if(graphics->hwnd) {
-        if(!GetClientRect(graphics->hwnd, &wnd_rect))
-            return GenericError;
-
-        rect->X = wnd_rect.left;
-        rect->Y = wnd_rect.top;
-        rect->Width = wnd_rect.right - wnd_rect.left;
-        rect->Height = wnd_rect.bottom - wnd_rect.top;
-    }else if (graphics->image){
-        stat = GdipGetImageBounds(graphics->image, rect, &unit);
-        if (stat == Ok && unit != UnitPixel)
-            FIXME("need to convert from unit %i\n", unit);
-    }else if (GetObjectType(graphics->hdc) == OBJ_MEMDC){
-        HBITMAP hbmp;
-        BITMAP bmp;
-
-        rect->X = 0;
-        rect->Y = 0;
-
-        hbmp = GetCurrentObject(graphics->hdc, OBJ_BITMAP);
-        if (hbmp && GetObjectW(hbmp, sizeof(bmp), &bmp))
-        {
-            rect->Width = bmp.bmWidth;
-            rect->Height = bmp.bmHeight;
-        }
-        else
-        {
-            /* FIXME: ??? */
-            rect->Width = 1;
-            rect->Height = 1;
-        }
-    }else{
-        rect->X = 0;
-        rect->Y = 0;
-        rect->Width = GetDeviceCaps(graphics->hdc, HORZRES);
-        rect->Height = GetDeviceCaps(graphics->hdc, VERTRES);
-    }
-
-    return stat;
-}
-
 static GpStatus get_clip_hrgn(GpGraphics *graphics, HRGN *hrgn)
 {
     GpRegion *rgn;
     GpMatrix transform;
     GpStatus stat;
-    GpRectF bounds;
-    RECT gdi_bounds;
     BOOL identity;
 
-    stat = get_graphics_device_bounds(graphics, &bounds);
-
-    if (stat == Ok)
-    {
-        gdi_bounds.left = floorf(bounds.X);
-        gdi_bounds.top = floorf(bounds.Y);
-        gdi_bounds.right = ceilf(bounds.X + bounds.Width);
-        gdi_bounds.bottom = ceilf(bounds.Y + bounds.Height);
-
-        stat = get_graphics_transform(graphics, WineCoordinateSpaceGdiDevice, CoordinateSpaceDevice, &transform);
-    }
+    stat = get_graphics_transform(graphics, WineCoordinateSpaceGdiDevice, CoordinateSpaceDevice, &transform);
 
     if (stat == Ok)
         stat = GdipIsMatrixIdentity(&transform, &identity);
@@ -492,7 +433,7 @@
             stat = GdipTransformRegion(rgn, &transform);
 
         if (stat == Ok)
-            stat = get_region_hrgn(&rgn->node, &gdi_bounds, hrgn);
+            stat = GdipGetRegionHRgn(rgn, NULL, hrgn);
 
         GdipDeleteRegion(rgn);
     }
@@ -550,27 +491,6 @@
     return Ok;
 }
 
-static void blend_32bppARGB(UINT width, UINT height, BYTE *dst_bits,
-    INT dst_stride, const BYTE *src_bits, INT src_stride)
-{
-    INT x, y;
-    for (y = 0; y < height; y++)
-    {
-        const BYTE *src = src_bits + y * src_stride;
-        BYTE *dst = dst_bits + y * dst_stride;
-        for (x = 0; x < width; x++)
-        {
-            BYTE alpha = src[3];
-            /* blend to white background */
-            *dst++ = (*src++ * alpha + 127) / 255 + (255 - alpha);
-            *dst++ = (*src++ * alpha + 127) / 255 + (255 - alpha);
-            *dst++ = (*src++ * alpha + 127) / 255 + (255 - alpha);
-            *dst++ = 255;
-            src++;
-        }
-    }
-}
-
 static GpStatus alpha_blend_hdc_pixels(GpGraphics *graphics, INT dst_x, INT dst_y,
     const BYTE *src, INT src_width, INT src_height, INT src_stride, PixelFormat fmt)
 {
@@ -599,12 +519,10 @@
     if(!hbitmap || !temp_bits)
         goto done;
 
-    if (graphics->hdc &&
+    if ((graphics->hdc &&
          GetDeviceCaps(graphics->hdc, TECHNOLOGY) == DT_RASPRINTER &&
-         GetDeviceCaps(graphics->hdc, SHADEBLENDCAPS) == SB_NONE)
-        blend_32bppARGB(src_width, src_height, temp_bits,
-                        4 * src_width, src, src_stride);
-    else if (fmt & PixelFormatPAlpha)
+         GetDeviceCaps(graphics->hdc, SHADEBLENDCAPS) == SB_NONE) ||
+            fmt & PixelFormatPAlpha)
         memcpy(temp_bits, src, src_width * src_height * 4);
     else
         convert_32bppARGB_to_32bppPARGB(src_width, src_height, temp_bits,
@@ -639,23 +557,21 @@
         if (!hrgn)
             return OutOfMemory;
 
+        stat = get_clip_hrgn(graphics, &visible_rgn);
+        if (stat != Ok)
+        {
+            DeleteObject(hrgn);
+            return stat;
+        }
+
+        if (visible_rgn)
+        {
+            CombineRgn(hrgn, hrgn, visible_rgn, RGN_AND);
+            DeleteObject(visible_rgn);
+        }
+
         if (hregion)
             CombineRgn(hrgn, hrgn, hregion, RGN_AND);
-        else
-        {
-            stat = get_clip_hrgn(graphics, &visible_rgn);
-            if (stat != Ok)
-            {
-                DeleteObject(hrgn);
-                return stat;
-            }
-
-            if (visible_rgn)
-            {
-                CombineRgn(hrgn, hrgn, visible_rgn, RGN_AND);
-                DeleteObject(visible_rgn);
-            }
-        }
 
         size = GetRegionData(hrgn, 0, NULL);
 
@@ -700,30 +616,27 @@
         if (stat != Ok)
             return stat;
 
+        stat = get_clip_hrgn(graphics, &hrgn);
+
+        if (stat != Ok)
+        {
+            gdi_dc_release(graphics, hdc);
+            return stat;
+        }
+
         save = SaveDC(hdc);
 
+        ExtSelectClipRgn(hdc, hrgn, RGN_COPY);
+
         if (hregion)
-            ExtSelectClipRgn(hdc, hregion, RGN_COPY);
-        else
-        {
-            stat = get_clip_hrgn(graphics, &hrgn);
-
-            if (stat != Ok)
-            {
-                RestoreDC(hdc, save);
-                gdi_dc_release(graphics, hdc);
-                return stat;
-            }
-
-            ExtSelectClipRgn(hdc, hrgn, RGN_COPY);
-
-            DeleteObject(hrgn);
-        }
+            ExtSelectClipRgn(hdc, hregion, RGN_AND);
 
         stat = alpha_blend_hdc_pixels(graphics, dst_x, dst_y, src, src_width,
             src_height, src_stride, fmt);
 
         RestoreDC(hdc, save);
+
+        DeleteObject(hrgn);
 
         gdi_dc_release(graphics, hdc);
 
@@ -2318,6 +2231,53 @@
     graphics->origin_y = container->origin_y;
 
     return Ok;
+}
+
+static GpStatus get_graphics_device_bounds(GpGraphics* graphics, GpRectF* rect)
+{
+    RECT wnd_rect;
+    GpStatus stat=Ok;
+    GpUnit unit;
+
+    if(graphics->hwnd) {
+        if(!GetClientRect(graphics->hwnd, &wnd_rect))
+            return GenericError;
+
+        rect->X = wnd_rect.left;
+        rect->Y = wnd_rect.top;
+        rect->Width = wnd_rect.right - wnd_rect.left;
+        rect->Height = wnd_rect.bottom - wnd_rect.top;
+    }else if (graphics->image){
+        stat = GdipGetImageBounds(graphics->image, rect, &unit);
+        if (stat == Ok && unit != UnitPixel)
+            FIXME("need to convert from unit %i\n", unit);
+    }else if (GetObjectType(graphics->hdc) == OBJ_MEMDC){
+        HBITMAP hbmp;
+        BITMAP bmp;
+
+        rect->X = 0;
+        rect->Y = 0;
+
+        hbmp = GetCurrentObject(graphics->hdc, OBJ_BITMAP);
+        if (hbmp && GetObjectW(hbmp, sizeof(bmp), &bmp))
+        {
+            rect->Width = bmp.bmWidth;
+            rect->Height = bmp.bmHeight;
+        }
+        else
+        {
+            /* FIXME: ??? */
+            rect->Width = 1;
+            rect->Height = 1;
+        }
+    }else{
+        rect->X = 0;
+        rect->Y = 0;
+        rect->Width = GetDeviceCaps(graphics->hdc, HORZRES);
+        rect->Height = GetDeviceCaps(graphics->hdc, VERTRES);
+    }
+
+    return stat;
 }
 
 static GpStatus get_graphics_bounds(GpGraphics* graphics, GpRectF* rect)
@@ -3400,11 +3360,7 @@
                 GdipSetMatrixElements(&dst_to_src, m11, m12, m21, m22, mdx, mdy);
 
                 stat = GdipInvertMatrix(&dst_to_src);
-                if (stat != Ok)
-                {
-                    free(src_data);
-                    return stat;
-                }
+                if (stat != Ok) return stat;
 
                 dst_stride = sizeof(ARGB) * (dst_area.right - dst_area.left);
                 x_dx = dst_to_src.matrix[0];
@@ -3478,7 +3434,7 @@
             if (bitmap->format == PixelFormat16bppRGB555 ||
                 bitmap->format == PixelFormat24bppRGB)
                 dst_format = bitmap->format;
-            else if (bitmap->image.flags & ImageFlagsHasAlpha)
+            else if (bitmap->format & (PixelFormatAlpha|PixelFormatPAlpha))
                 dst_format = PixelFormat32bppPARGB;
             else
                 dst_format = PixelFormat32bppRGB;
@@ -3521,7 +3477,7 @@
 
             gdi_transform_acquire(graphics);
 
-            if (bitmap->image.flags & ImageFlagsHasAlpha)
+            if (bitmap->format & (PixelFormatAlpha|PixelFormatPAlpha))
             {
                 gdi_alpha_blend(graphics, pti[0].x, pti[0].y, pti[1].x - pti[0].x, pti[2].y - pti[0].y,
                                 src_hdc, srcx, srcy, srcwidth, srcheight);
@@ -4795,84 +4751,6 @@
     return ret;
 }
 
-static GpStatus get_clipped_region_hrgn(GpGraphics* graphics, GpRegion* region, HRGN *hrgn)
-{
-    GpStatus status;
-    GpRegion *tmp_region, *device_region;
-    GpMatrix gdip_transform, gdi_transform;
-    BOOL gdip_identity, gdi_identity;
-    GpRectF device_bounds;
-
-    status = get_graphics_device_bounds(graphics, &device_bounds);
-
-    if (status == Ok)
-        status = get_graphics_transform(graphics, WineCoordinateSpaceGdiDevice, CoordinateSpaceWorld, &gdip_transform);
-
-    if (status == Ok)
-        status = GdipIsMatrixIdentity(&gdip_transform, &gdip_identity);
-
-    if (status == Ok)
-        status = get_graphics_transform(graphics, WineCoordinateSpaceGdiDevice, CoordinateSpaceDevice, &gdi_transform);
-
-    if (status == Ok)
-        status = GdipIsMatrixIdentity(&gdi_transform, &gdi_identity);
-
-    if (status == Ok)
-        status = GdipCreateRegionRect(&device_bounds, &device_region);
-
-    if (status == Ok)
-    {
-        if (gdip_identity)
-            status = GdipCombineRegionRegion(device_region, region, CombineModeIntersect);
-        else
-        {
-            status = GdipCloneRegion(region, &tmp_region);
-
-            if (status == Ok)
-            {
-                status = GdipTransformRegion(tmp_region, &gdip_transform);
-
-                if (status == Ok)
-                    GdipCombineRegionRegion(device_region, tmp_region, CombineModeIntersect);
-
-                GdipDeleteRegion(tmp_region);
-            }
-        }
-
-        if (status == Ok)
-        {
-            if (gdi_identity)
-                status = GdipCombineRegionRegion(device_region, graphics->clip, CombineModeIntersect);
-            else
-            {
-                status = GdipCloneRegion(graphics->clip, &tmp_region);
-
-                if (status == Ok)
-                {
-                    status = GdipTransformRegion(tmp_region, &gdi_transform);
-
-                    if (status == Ok)
-                        GdipCombineRegionRegion(device_region, tmp_region, CombineModeIntersect);
-
-                    GdipDeleteRegion(tmp_region);
-                }
-            }
-        }
-
-        if (status == Ok)
-            status = GdipGetRegionHRgn(device_region, NULL, hrgn);
-
-        GdipDeleteRegion(device_region);
-    }
-
-    if (status == Ok && graphics->gdi_clip)
-    {
-        CombineRgn(*hrgn, *hrgn, graphics->gdi_clip, RGN_AND);
-    }
-
-    return status;
-}
-
 static GpStatus GDI32_GdipFillRegion(GpGraphics* graphics, GpBrush* brush,
     GpRegion* region)
 {
@@ -4893,11 +4771,18 @@
     EndPath(hdc);
 
     hrgn = NULL;
-    status = get_clipped_region_hrgn(graphics, region, &hrgn);
+    status = get_clip_hrgn(graphics, &hrgn);
     if (status != Ok)
         goto end;
 
     ExtSelectClipRgn(hdc, hrgn, RGN_COPY);
+    DeleteObject(hrgn);
+
+    status = GdipGetRegionHRgn(region, graphics, &hrgn);
+    if (status != Ok)
+        goto end;
+
+    ExtSelectClipRgn(hdc, hrgn, RGN_AND);
     DeleteObject(hrgn);
 
     if (GetClipBox(hdc, &rc) != NULLREGION)
@@ -4920,6 +4805,9 @@
     GpRegion* region)
 {
     GpStatus stat;
+    GpRegion *temp_region;
+    GpMatrix world_to_device;
+    GpRectF graphics_bounds;
     DWORD *pixel_data;
     HRGN hregion;
     RECT bound_rect;
@@ -4931,7 +4819,27 @@
     stat = gdi_transform_acquire(graphics);
 
     if (stat == Ok)
-        stat = get_clipped_region_hrgn(graphics, region, &hregion);
+        stat = get_graphics_device_bounds(graphics, &graphics_bounds);
+
+    if (stat == Ok)
+        stat = GdipCloneRegion(region, &temp_region);
+
+    if (stat == Ok)
+    {
+        stat = get_graphics_transform(graphics, WineCoordinateSpaceGdiDevice,
+            CoordinateSpaceWorld, &world_to_device);
+
+        if (stat == Ok)
+            stat = GdipTransformRegion(temp_region, &world_to_device);
+
+        if (stat == Ok)
+            stat = GdipCombineRegionRect(temp_region, &graphics_bounds, CombineModeIntersect);
+
+        if (stat == Ok)
+            stat = GdipGetRegionHRgn(temp_region, NULL, &hregion);
+
+        GdipDeleteRegion(temp_region);
+    }
 
     if (stat == Ok && GetRgnBox(hregion, &bound_rect) == NULLREGION)
     {
@@ -5454,7 +5362,6 @@
     DWORD string_codepages;
     WORD *glyph_indices;
     HRESULT hr;
-    GpStatus stat;
 
     list_init(&info->font_link_info.sections);
     info->font_link_info.base_font = base_font;
@@ -5493,13 +5400,10 @@
             if (SUCCEEDED(hr))
             {
                 old_font = SelectObject(info->hdc, map_hfont);
-                stat = GdipCreateFontFromDC(info->hdc, &gpfont);
+                GdipCreateFontFromDC(info->hdc, &gpfont);
                 SelectObject(info->hdc, old_font);
                 IMLangFontLink_ReleaseFont(iMLFL, map_hfont);
-                if (stat == Ok)
-                    section->font = gpfont;
-                else
-                    section->font = (GpFont *)base_font;
+                section->font = gpfont;
             }
             else
                 section->font = (GpFont *)base_font;
@@ -5999,6 +5903,15 @@
     if(!graphics || !string || !font || !rect || !bounds)
         return InvalidParameter;
 
+    if (length == 1 && string[0] == '\n')
+    {
+        /* Proton hack for SpriteFontX class used by TouHou Makuka Sai.
+         * Returned size is passed to Bitmap constructor, but we currently measure "\n" as zero size. */
+        char const *sgi = getenv("SteamGameId");
+        if (sgi && (!strcmp(sgi, "882710") || !strcmp(sgi, "1031480")))
+            string = L" ";
+    }
+
     if(!has_gdi_dc(graphics))
     {
         hdc = temp_hdc = CreateCompatibleDC(0);
@@ -6110,7 +6023,6 @@
 
         GetOutlineTextMetricsW(info->hdc, sizeof(otm), &otm);
 
-        position.X = args->x + info->bounds->X / args->rel_width;
         underline_height = otm.otmsUnderscoreSize / args->rel_height;
         underline_y = position.Y - otm.otmsUnderscorePosition / args->rel_height - underline_height / 2;
 
