diff --git a/dlls/iphlpapi/iphlpapi_main.c b/dlls/iphlpapi/iphlpapi_main.c
--- a/dlls/iphlpapi/iphlpapi_main.c
+++ b/dlls/iphlpapi/iphlpapi_main.c
@@ -38,9 +38,11 @@
 #include "ip2string.h"
 #include "netiodef.h"
 #include "icmpapi.h"
+#include "psapi.h"
 
 #include "wine/nsi.h"
 #include "wine/debug.h"
+#include "wine/heap.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(iphlpapi);
 
@@ -511,7 +513,7 @@
 void WINAPI FreeMibTable( void *ptr )
 {
     TRACE( "(%p)\n", ptr );
-    HeapFree( GetProcessHeap(), 0, ptr );
+    heap_free( ptr );
 }
 
 /******************************************************************
@@ -563,7 +565,7 @@
 
     if (count)
     {
-        *servers = calloc( count, sizeof(**servers) );
+        *servers = heap_alloc_zero( count * sizeof(**servers) );
         if (!*servers) return 0;
         for (i = 0; i < count; i++)
         {
@@ -725,7 +727,7 @@
     info[-1].Next = NULL;
 
 err:
-    free( wins_servers );
+    heap_free( wins_servers );
     NsiFreeTable( fwd_keys, NULL, NULL, NULL );
     NsiFreeTable( uni_keys, uni_rw, NULL, NULL );
     NsiFreeTable( if_keys, if_rw, NULL, if_stat );
@@ -734,7 +736,7 @@
 
 static void address_entry_free( void *ptr, ULONG offset, void *ctxt )
 {
-    HeapFree( GetProcessHeap(), 0, ptr );
+    heap_free( ptr );
 }
 
 static void address_entry_size( void *ptr, ULONG offset, void *ctxt )
@@ -816,9 +818,9 @@
     {
         address_lists_iterate( aa, address_entry_free, NULL );
 
-        HeapFree( GetProcessHeap(), 0, aa->DnsSuffix );
-    }
-    HeapFree( GetProcessHeap(), 0, info );
+        heap_free( aa->DnsSuffix );
+    }
+    heap_free( info );
 }
 
 static ULONG adapters_addresses_size( IP_ADAPTER_ADDRESSES *info )
@@ -852,7 +854,7 @@
     IP_ADAPTER_ADDRESSES **sorted;
     ULONG i = 0;
 
-    if (!(sorted = malloc( count * sizeof(*sorted) ))) return NULL;
+    if (!(sorted = heap_alloc( count * sizeof(*sorted) ))) return NULL;
 
     while (src)
     {
@@ -974,7 +976,7 @@
             key6 = (struct nsi_ipv6_unicast_key *)key + i;
             luid = (family == AF_INET) ? &key4->luid : &key6->luid;
             if (luid->Value != aa->Luid.Value) continue;
-            addr = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*addr) + sockaddr_size );
+            addr = heap_alloc_zero( sizeof(*addr) + sockaddr_size );
             if (!addr)
             {
                 err = ERROR_NOT_ENOUGH_MEMORY;
@@ -1080,7 +1082,7 @@
 
                 if (sockaddr.si_family)
                 {
-                    gw = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*gw) + sockaddr_size );
+                    gw = heap_alloc_zero( sizeof(*gw) + sockaddr_size );
                     if (!gw)
                     {
                         err = ERROR_NOT_ENOUGH_MEMORY;
@@ -1121,7 +1123,7 @@
 
                 if (sockaddr.si_family)
                 {
-                    prefix = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*prefix) + sockaddr_size );
+                    prefix = heap_alloc_zero( sizeof(*prefix) + sockaddr_size );
                     if (!prefix)
                     {
                         err = ERROR_NOT_ENOUGH_MEMORY;
@@ -1191,8 +1193,8 @@
             {
                 err = DnsQueryConfig( query, 0, name, NULL, servers, &size );
                 if (err != ERROR_MORE_DATA) break;
-                if (servers != (DNS_ADDR_ARRAY *)buf) free( servers );
-                servers = malloc( size );
+                if (servers != (DNS_ADDR_ARRAY *)buf) heap_free( servers );
+                servers = heap_alloc( size );
                 if (!servers)
                 {
                     err = ERROR_NOT_ENOUGH_MEMORY;
@@ -1207,7 +1209,7 @@
                     sockaddr_len = servers->AddrArray[i].Data.DnsAddrUserDword[0];
                     if (sockaddr_len > sizeof(servers->AddrArray[i].MaxSa))
                         sockaddr_len = sizeof(servers->AddrArray[i].MaxSa);
-                    dns = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(*dns) + sockaddr_len );
+                    dns = heap_alloc_zero( sizeof(*dns) + sockaddr_len );
                     if (!dns)
                     {
                         err = ERROR_NOT_ENOUGH_MEMORY;
@@ -1221,23 +1223,23 @@
                     next = &dns->Next;
                 }
             }
-            if (servers != (DNS_ADDR_ARRAY *)buf) free( servers );
+            if (servers != (DNS_ADDR_ARRAY *)buf) heap_free( servers );
             if (err) return err;
         }
 
-        aa->DnsSuffix = HeapAlloc( GetProcessHeap(), 0, MAX_DNS_SUFFIX_STRING_LENGTH * sizeof(WCHAR) );
+        aa->DnsSuffix = heap_alloc( MAX_DNS_SUFFIX_STRING_LENGTH * sizeof(WCHAR) );
         if (!aa->DnsSuffix) return ERROR_NOT_ENOUGH_MEMORY;
         aa->DnsSuffix[0] = '\0';
 
         if (!DnsQueryConfig( DnsConfigSearchList, 0, name, NULL, NULL, &size ) &&
-            (search = malloc( size )))
+            (search = heap_alloc( size )))
         {
             if (!DnsQueryConfig( DnsConfigSearchList, 0, name, NULL, search, &size ) &&
                 search[0] && wcslen( search ) < MAX_DNS_SUFFIX_STRING_LENGTH)
             {
                 wcscpy( aa->DnsSuffix, search );
             }
-            free( search );
+            heap_free( search );
         }
 
         aa = aa->Next;
@@ -1265,7 +1267,7 @@
     needed = *count * (sizeof(*aa) + ((CHARS_IN_GUID + 1) & ~1) + sizeof(stat->descr.String));
     needed += *count * sizeof(rw->alias.String); /* GAA_FLAG_SKIP_FRIENDLY_NAME is ignored */
 
-    aa = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, needed );
+    aa = heap_alloc_zero( needed );
     if (!aa)
     {
         err = ERROR_NOT_ENOUGH_MEMORY;
@@ -1345,7 +1347,7 @@
         else
         {
             adapters_addresses_copy( aa, sorted, count );
-            free( sorted );
+            heap_free( sorted );
         }
     }
 
@@ -1391,20 +1393,14 @@
 DWORD WINAPI GetBestInterfaceEx( struct sockaddr *dst, DWORD *best_index )
 {
     SOCKADDR_INET best_address;
-    SOCKADDR_INET dst_address;
     MIB_IPFORWARD_ROW2 row;
     DWORD ret;
 
     TRACE( "dst %p, best_index %p\n", dst, best_index );
 
     if (!dst || !best_index) return ERROR_INVALID_PARAMETER;
-    if (dst->sa_family != AF_INET && dst->sa_family != AF_INET6) return ERROR_INVALID_PARAMETER;
-
-    dst_address.si_family = dst->sa_family;
-    if (dst->sa_family == AF_INET6) dst_address.Ipv6 = *(struct sockaddr_in6 *)dst;
-    else dst_address.Ipv4 = *(struct sockaddr_in *)dst;
-
-    ret = GetBestRoute2( NULL, 0, NULL, &dst_address, 0, &row, &best_address );
+
+    ret = GetBestRoute2( NULL, 0, NULL, (const SOCKADDR_INET *)dst, 0, &row, &best_address );
     if (!ret) *best_index = row.InterfaceIndex;
 
     TRACE( "returning %ld\n", ret );
@@ -1882,7 +1878,7 @@
 
     size = FIELD_OFFSET( MIB_IF_TABLE2, Table[count] );
 
-    if (!(*table = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, size )))
+    if (!(*table = heap_alloc_zero( size )))
     {
         err = ERROR_OUTOFMEMORY;
         goto err;
@@ -2298,7 +2294,7 @@
     }
 
     size = FIELD_OFFSET(MIB_IPFORWARD_TABLE2, Table[ count[0] + count[1] ]);
-    *table = HeapAlloc( GetProcessHeap(), 0, size );
+    *table = heap_alloc( size );
     if (!*table)
     {
         err = ERROR_NOT_ENOUGH_MEMORY;
@@ -2515,7 +2511,7 @@
     }
 
     size = FIELD_OFFSET(MIB_IPNET_TABLE2, Table[ count[0] + count[1] ]);
-    *table = HeapAlloc( GetProcessHeap(), 0, size );
+    *table = heap_alloc( size );
     if (!*table)
     {
         err = ERROR_NOT_ENOUGH_MEMORY;
@@ -2660,8 +2656,8 @@
         }
         if (!err) break;
 
-        if ((char *)servers != buf) free( servers );
-        servers = malloc( array_len );
+        if ((char *)servers != buf) heap_free( servers );
+        servers = heap_alloc( array_len );
         if (!servers)
         {
             err = ERROR_NOT_ENOUGH_MEMORY;
@@ -2680,7 +2676,7 @@
     }
 
 err:
-    if ((char *)servers != buf) free( servers );
+    if ((char *)servers != buf) heap_free( servers );
     return err;
 }
 
@@ -3271,6 +3267,68 @@
     return get_extended_tcp_table( table, size, sort, family, table_class );
 }
 
+static DWORD get_owner_module_from_pid( TCPIP_OWNER_MODULE_INFO_CLASS class, void *buffer, DWORD *size, DWORD pid )
+{
+    WCHAR path[(MAX_PATH + 1) * sizeof(WCHAR)], *name;
+    TCPIP_OWNER_MODULE_BASIC_INFO *info = buffer;
+    DWORD len, ret_size;
+    HANDLE process;
+
+    if (class != TCPIP_OWNER_MODULE_INFO_BASIC)
+    {
+        FIXME( "Unsupported class %d.\n", class );
+        return ERROR_INVALID_PARAMETER;
+    }
+
+    if (!size) return ERROR_INVALID_PARAMETER;
+    if (!pid) return ERROR_NOT_FOUND;
+    if (!(process = OpenProcess( PROCESS_QUERY_LIMITED_INFORMATION, FALSE, pid ))) return GetLastError();
+    len = GetModuleFileNameExW( process, NULL, path, ARRAY_SIZE(path) );
+    CloseHandle( process );
+    if (!len) return ERROR_NOT_FOUND;
+    if (len == ARRAY_SIZE(path)) return ERROR_PATH_NOT_FOUND;
+    if (!(name = wcsrchr( path, '\\' ))) return ERROR_PATH_NOT_FOUND;
+    ++name;
+    len = wcslen( name ) + 1;
+    ret_size = sizeof(*info) + (len + wcslen( path ) + 1) * sizeof(WCHAR);
+    if (*size < ret_size)
+    {
+        *size = ret_size;
+        return ERROR_INSUFFICIENT_BUFFER;
+    }
+    info->pModuleName = (WCHAR *)(info + 1);
+    info->pModulePath = info->pModuleName + len;
+    memcpy( info->pModuleName, name, len * sizeof(WCHAR) );
+    wcscpy( info->pModulePath, path );
+    return ERROR_SUCCESS;
+}
+
+/******************************************************************
+ *    GetOwnerModuleFromTcpEntry (IPHLPAPI.@)
+ */
+DWORD WINAPI GetOwnerModuleFromTcpEntry( PMIB_TCPROW_OWNER_MODULE entry, TCPIP_OWNER_MODULE_INFO_CLASS class,
+                                         void *buffer, DWORD *size )
+{
+    TRACE( "entry %p, class %d, buffer %p, size %p.\n", entry, class, buffer, size );
+
+    if (!entry) return ERROR_INVALID_PARAMETER;
+
+    return get_owner_module_from_pid( class, buffer, size, entry->dwOwningPid );
+}
+
+/******************************************************************
+ *    GetOwnerModuleFromTcp6Entry (IPHLPAPI.@)
+ */
+DWORD WINAPI GetOwnerModuleFromTcp6Entry( PMIB_TCP6ROW_OWNER_MODULE entry, TCPIP_OWNER_MODULE_INFO_CLASS class,
+                                         void *buffer, DWORD *size )
+{
+    TRACE( "entry %p, class %d, buffer %p, size %p.\n", entry, class, buffer, size );
+
+    if (!entry) return ERROR_INVALID_PARAMETER;
+
+    return get_owner_module_from_pid( class, buffer, size, entry->dwOwningPid );
+}
+
 /******************************************************************
  *    GetTcpTable (IPHLPAPI.@)
  *
@@ -3752,7 +3810,7 @@
     }
 
     size = FIELD_OFFSET(MIB_UNICASTIPADDRESS_TABLE, Table[ count[0] + count[1] ]);
-    *table = HeapAlloc( GetProcessHeap(), 0, size );
+    *table = heap_alloc( size );
     if (!*table)
     {
         err = ERROR_NOT_ENOUGH_MEMORY;
@@ -3787,7 +3845,7 @@
     if (!table || (family != AF_INET && family != AF_INET6 && family != AF_UNSPEC))
         return ERROR_INVALID_PARAMETER;
 
-    *table = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(MIB_ANYCASTIPADDRESS_TABLE));
+    *table = heap_alloc_zero(sizeof(MIB_ANYCASTIPADDRESS_TABLE));
     if (!*table) return ERROR_NOT_ENOUGH_MEMORY;
     (*table)->NumEntries = 0;
     return NO_ERROR;
@@ -4136,19 +4194,6 @@
     return ERROR_CALL_NOT_IMPLEMENTED;
 }
 
-/***********************************************************************
- *    GetPerTcp6ConnectionEStats (IPHLPAPI.@)
- */
-ULONG WINAPI GetPerTcp6ConnectionEStats(MIB_TCP6ROW *row, TCP_ESTATS_TYPE stats, UCHAR *rw, ULONG rw_version,
-                                        ULONG rw_size, UCHAR *ro_static, ULONG ro_static_version,
-                                        ULONG ro_static_size, UCHAR *ro_dynamic, ULONG ro_dynamic_version,
-                                        ULONG ro_dynamic_size)
-{
-    FIXME( "(%p, %d, %p, %ld, %ld, %p, %ld, %ld, %p, %ld, %ld): stub\n", row, stats, rw, rw_version, rw_size,
-           ro_static, ro_static_version, ro_static_size, ro_dynamic, ro_dynamic_version, ro_dynamic_size );
-    return ERROR_CALL_NOT_IMPLEMENTED;
-}
-
 /******************************************************************
  *    SetPerTcpConnectionEStats (IPHLPAPI.@)
  */
@@ -4160,16 +4205,6 @@
   return ERROR_NOT_SUPPORTED;
 }
 
-/******************************************************************
- *    SetPerTcp6ConnectionEStats (IPHLPAPI.@)
- */
-DWORD WINAPI SetPerTcp6ConnectionEStats(MIB_TCP6ROW *row, TCP_ESTATS_TYPE state, BYTE *rw,
-                                        ULONG version, ULONG size, ULONG offset)
-{
-    FIXME("(row %p, state %d, rw %p, version %lu, size %lu, offset %lu): stub\n",
-          row, state, rw, version, size, offset);
-    return ERROR_NOT_SUPPORTED;
-}
 
 /******************************************************************
  *    UnenableRouter (IPHLPAPI.@)
@@ -4626,21 +4661,21 @@
                                       (void **)&dyn, sizeof(*dyn), NULL, 0, &count, 0 );
         if (err)
         {
-            HeapFree( GetProcessHeap(), 0, *table );
+            heap_free( *table );
             return err;
         }
         total_count += count;
-        new_alloc = HeapAlloc( GetProcessHeap(), HEAP_ZERO_MEMORY, offsetof(MIB_IPINTERFACE_TABLE, Table[total_count]) );
+        new_alloc = heap_alloc_zero( offsetof(MIB_IPINTERFACE_TABLE, Table[total_count]) );
         if (!new_alloc)
         {
-            HeapFree( GetProcessHeap(), 0, *table );
+            heap_free( *table );
             NsiFreeTable( keys, rw, dyn, NULL );
             return ERROR_NOT_ENOUGH_MEMORY;
         }
         if (*table)
         {
             memcpy( new_alloc, *table, offsetof(MIB_IPINTERFACE_TABLE, Table[(*table)->NumEntries]) );
-            HeapFree( GetProcessHeap(), 0, *table );
+            free( *table );
         }
         *table = new_alloc;
         for (i = 0; i < count; ++i)
@@ -4679,6 +4714,7 @@
     fill_ip_interface_table_row( row->Family, row, &key, &rw, &dyn );
     return ERROR_SUCCESS;
 }
+
 
 static BOOL match_ip_address_with_prefix( const SOCKADDR_INET *addr, const IP_ADDRESS_PREFIX *pfx )
 {
@@ -5041,7 +5077,7 @@
     struct icmp_apc_ctxt *ctx = (struct icmp_apc_ctxt *)context;
 
     ctx->apc_routine( ctx->apc_ctxt, &ctx->iosb, 0 );
-    free( ctx );
+    heap_free( ctx );
 }
 
 static void CALLBACK icmp_iocp_callback( DWORD error, DWORD count, OVERLAPPED *ovr )
@@ -5061,7 +5097,7 @@
         CloseHandle( thread );
         if (ret) return;
     }
-    free( ctx );
+    heap_free( ctx );
 }
 
 /***********************************************************************
@@ -5077,7 +5113,7 @@
     data = (struct icmp_handle_data *)handle;
 
     CloseHandle( data->nsi_device );
-    HeapFree( GetProcessHeap(), 0, data );
+    heap_free( data );
     return TRUE;
 }
 
@@ -5086,7 +5122,7 @@
  */
 HANDLE WINAPI IcmpCreateFile( void )
 {
-    struct icmp_handle_data *data = HeapAlloc( GetProcessHeap(), 0, sizeof(*data) );
+    struct icmp_handle_data *data = heap_alloc( sizeof(*data) );
 
     if (!data)
     {
@@ -5098,14 +5134,14 @@
                                     FILE_FLAG_OVERLAPPED, NULL );
     if (data->nsi_device == INVALID_HANDLE_VALUE)
     {
-        HeapFree( GetProcessHeap(), 0, data );
+        heap_free( data );
         return INVALID_HANDLE_VALUE;
     }
     if (!BindIoCompletionCallback( data->nsi_device, icmp_iocp_callback, 0 ))
     {
         ERR( "BindIoCompletionCallback failed.\n" );
         CloseHandle( data->nsi_device );
-        HeapFree( GetProcessHeap(), 0, data );
+        heap_free( data );
         return INVALID_HANDLE_VALUE;
     }
     return (HANDLE)data;
@@ -5163,7 +5199,7 @@
 
     opt_size = opts ? (opts->OptionsSize + 3) & ~3 : 0;
     in_size = FIELD_OFFSET(struct nsiproxy_icmp_echo, data[opt_size + request_size]);
-    in = calloc( 1, in_size );
+    in = heap_alloc_zero( in_size );
 
     if (!in) return STATUS_NO_MEMORY;
 
@@ -5187,9 +5223,9 @@
 
     if (event || apc_routine)
     {
-        if (!(ctxt = malloc( sizeof(*ctxt) )))
-        {
-            free( in );
+        if (!(ctxt = heap_alloc( sizeof(*ctxt) )))
+        {
+            heap_free( in );
             return STATUS_NO_MEMORY;
         }
         iosb = &ctxt->iosb;
@@ -5202,8 +5238,8 @@
             if (!DuplicateHandle( GetCurrentProcess(), GetCurrentThread(), GetCurrentProcess(),
                              &ctxt->thread, 0, FALSE, DUPLICATE_SAME_ACCESS ))
             {
-                free( ctxt );
-                free( in );
+                heap_free( ctxt );
+                heap_free( in );
                 return STATUS_NO_MEMORY;
             }
         }
@@ -5224,9 +5260,9 @@
     if (ctxt && status != STATUS_PENDING)
     {
         if (ctxt->thread) CloseHandle( ctxt->thread );
-        free( ctxt );
-    }
-    free( in );
+        heap_free( ctxt );
+    }
+    heap_free( in );
     return status;
 }
 
