diff --git a/dlls/kernel32/tests/loader.c b/dlls/kernel32/tests/loader.c
--- a/dlls/kernel32/tests/loader.c
+++ b/dlls/kernel32/tests/loader.c
@@ -36,6 +36,16 @@
 
 #define ALIGN_SIZE(size, alignment) (((size) + ((ULONG_PTR)(alignment) - 1)) & ~(((ULONG_PTR)(alignment) - 1)))
 
+struct PROCESS_BASIC_INFORMATION_PRIVATE
+{
+    NTSTATUS  ExitStatus;
+    PPEB      PebBaseAddress;
+    DWORD_PTR AffinityMask;
+    DWORD_PTR BasePriority;
+    ULONG_PTR UniqueProcessId;
+    ULONG_PTR InheritedFromUniqueProcessId;
+};
+
 static LONG *child_failures;
 static WORD cb_count, cb_count_sys;
 static DWORD page_size;
@@ -58,7 +68,7 @@
 static NTSTATUS (WINAPI *pNtFreeVirtualMemory)(HANDLE, PVOID *, SIZE_T *, ULONG);
 static NTSTATUS (WINAPI *pLdrLockLoaderLock)(ULONG, ULONG *, ULONG_PTR *);
 static NTSTATUS (WINAPI *pLdrUnlockLoaderLock)(ULONG, ULONG_PTR);
-static NTSTATUS (WINAPI *pLdrLoadDll)(LPCWSTR,DWORD,const UNICODE_STRING *,HMODULE*);
+static NTSTATUS (WINAPI *pLdrLoadDll)(LPCWSTR,DWORD *,const UNICODE_STRING *,HMODULE*);
 static NTSTATUS (WINAPI *pLdrUnloadDll)(HMODULE);
 static void (WINAPI *pRtlInitUnicodeString)(PUNICODE_STRING,LPCWSTR);
 static void (WINAPI *pRtlAcquirePebLock)(void);
@@ -78,7 +88,6 @@
 static BOOL (WINAPI *pWow64RevertWow64FsRedirection)(void *);
 static HMODULE (WINAPI *pLoadPackagedLibrary)(LPCWSTR lpwLibFileName, DWORD Reserved);
 static NTSTATUS  (WINAPI *pLdrRegisterDllNotification)(ULONG, PLDR_DLL_NOTIFICATION_FUNCTION, void *, void **);
-static NTSTATUS  (WINAPI *pLdrUnregisterDllNotification)(void *);
 
 static PVOID RVAToAddr(DWORD_PTR rva, HMODULE module)
 {
@@ -1632,7 +1641,7 @@
     addr1 = NULL;
     size = 0;
     status = pNtMapViewOfSection(hmap, GetCurrentProcess(), &addr1, 0, 0, &offset,
-                                 &size, ViewShare, 0, PAGE_READONLY);
+                                 &size, 1 /* ViewShare */, 0, PAGE_READONLY);
     ok(NT_SUCCESS(status), "NtMapViewOfSection error %lx\n", status);
     ok(addr1 != 0, "mapped address should be valid\n");
 
@@ -1650,7 +1659,7 @@
     addr2 = NULL;
     size = 0;
     status = pNtMapViewOfSection(hmap, GetCurrentProcess(), &addr2, 0, 0, &offset,
-                                 &size, ViewShare, 0, PAGE_READONLY);
+                                 &size, 1 /* ViewShare */, 0, PAGE_READONLY);
     ok(status == STATUS_IMAGE_NOT_AT_BASE, "expected STATUS_IMAGE_NOT_AT_BASE, got %lx\n", status);
     ok(addr2 != 0, "mapped address should be valid\n");
     ok(addr2 != addr1, "mapped addresses should be different\n");
@@ -2112,8 +2121,8 @@
 
 static void test_import_resolution(void)
 {
-    char temp_path[MAX_PATH];
-    char dll_name[MAX_PATH];
+    WCHAR temp_path[MAX_PATH];
+    WCHAR dll_name[MAX_PATH];
     DWORD dummy;
     void *expect, *tmp;
     char *str;
@@ -2122,6 +2131,7 @@
     HMODULE mod, mod2;
     NTSTATUS status;
     LARGE_INTEGER offset;
+    UNICODE_STRING name;
     struct imports
     {
         IMAGE_IMPORT_DESCRIPTOR descr[2];
@@ -2177,7 +2187,7 @@
     static const UCHAR entry_point_code[] = { 0x00 };
 #endif
 
-    for (test = 0; test < 7; test++)
+    for (test = 0; test < 8; test++)
     {
 #define DATA_RVA(ptr) (page_size + ((char *)(ptr) - (char *)&data))
 #ifdef _WIN64
@@ -2246,10 +2256,10 @@
         nt.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_BASERELOC].Size = data.rel.reloc.SizeOfBlock;
         nt.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_BASERELOC].VirtualAddress = DATA_RVA(&data.rel);
 
-        GetTempPathA(MAX_PATH, temp_path);
-        GetTempFileNameA(temp_path, "ldr", 0, dll_name);
-
-        hfile = CreateFileA(dll_name, GENERIC_WRITE, FILE_SHARE_READ, NULL, CREATE_ALWAYS, 0, 0);
+        GetTempPathW(MAX_PATH, temp_path);
+        GetTempFileNameW(temp_path, L"ldr", 0, dll_name);
+
+        hfile = CreateFileW(dll_name, GENERIC_WRITE, FILE_SHARE_READ, NULL, CREATE_ALWAYS, 0, 0);
         ok( hfile != INVALID_HANDLE_VALUE, "creation failed\n" );
 
         memset( &section, 0, sizeof(section) );
@@ -2265,7 +2275,7 @@
         WriteFile(hfile, &nt, sizeof(nt), &dummy, NULL);
         WriteFile(hfile, &section, sizeof(section), &dummy, NULL);
 
-        SetFilePointer( hfile, section.PointerToRawData, NULL, FILE_BEGIN );
+        SetFilePointer( hfile, section.PointerToRawData, NULL, SEEK_SET );
         WriteFile(hfile, &data, sizeof(data), &dummy, NULL);
 
         CloseHandle( hfile );
@@ -2273,7 +2283,7 @@
         switch (test)
         {
         case 0:  /* normal load */
-            mod = LoadLibraryA( dll_name );
+            mod = LoadLibraryW( dll_name );
             ok( mod != NULL, "failed to load err %lu\n", GetLastError() );
             if (!mod) break;
             ptr = (struct imports *)((char *)mod + page_size);
@@ -2287,8 +2297,20 @@
             check_tls_index(mod, ptr->tls_index != 9999);
             FreeLibrary( mod );
             break;
+        case 7:
         case 1:  /* load with DONT_RESOLVE_DLL_REFERENCES doesn't resolve imports */
-            mod = LoadLibraryExA( dll_name, 0, DONT_RESOLVE_DLL_REFERENCES );
+            if (test == 7)
+            {
+                DWORD load_flags = LDR_DONT_RESOLVE_REFS;
+
+                pRtlInitUnicodeString( &name, dll_name );
+                status = pLdrLoadDll( NULL, &load_flags, &name, &mod );
+                ok( !status, "got %#lx.\n", status );
+            }
+            else
+            {
+                mod = LoadLibraryExW( dll_name, 0, DONT_RESOLVE_DLL_REFERENCES );
+            }
             ok( mod != NULL, "failed to load err %lu\n", GetLastError() );
             if (!mod) break;
             ptr = (struct imports *)((char *)mod + page_size);
@@ -2296,7 +2318,7 @@
                 (void *)ptr->thunks[0].u1.Function, data.module, data.function.name );
             ok( ptr->tls_index == 9999, "wrong tls index %d\n", ptr->tls_index );
 
-            mod2 = LoadLibraryA( dll_name );
+            mod2 = LoadLibraryW( dll_name );
             ok( mod2 == mod, "loaded twice %p / %p\n", mod, mod2 );
             ok( ptr->thunks[0].u1.Function == 0xdeadbeef, "thunk resolved to %p for %s.%s\n",
                 (void *)ptr->thunks[0].u1.Function, data.module, data.function.name );
@@ -2306,7 +2328,7 @@
             FreeLibrary( mod );
             break;
         case 2:  /* load without IMAGE_FILE_DLL doesn't resolve imports */
-            mod = LoadLibraryA( dll_name );
+            mod = LoadLibraryW( dll_name );
             ok( mod != NULL, "failed to load err %lu\n", GetLastError() );
             if (!mod) break;
             ptr = (struct imports *)((char *)mod + page_size);
@@ -2317,7 +2339,7 @@
             FreeLibrary( mod );
             break;
         case 3:  /* load with tls init function */
-            mod = LoadLibraryA( dll_name );
+            mod = LoadLibraryW( dll_name );
             ok( mod != NULL, "failed to load err %lu\n", GetLastError() );
             if (!mod) break;
             ptr = (struct imports *)((char *)mod + page_size);
@@ -2345,7 +2367,7 @@
         case 4:  /* map with ntdll */
         case 5:  /* map with ntdll, without IMAGE_FILE_DLL */
         case 6:  /* map with ntdll, without IMAGE_DLLCHARACTERISTICS_DYNAMIC_BASE */
-            hfile = CreateFileA(dll_name, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, 0);
+            hfile = CreateFileW(dll_name, GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, 0, 0);
             ok( hfile != INVALID_HANDLE_VALUE, "CreateFile failed err %lu\n", GetLastError() );
             mapping = CreateFileMappingA( hfile, NULL, SEC_IMAGE | PAGE_READONLY, 0, 0, NULL );
             CloseHandle( hfile );
@@ -2370,7 +2392,7 @@
             size = 0;
             offset.QuadPart = 0;
             status = pNtMapViewOfSection( mapping, GetCurrentProcess(), (void **)&mod, 0, 0, &offset,
-                                          &size, ViewShare, 0, PAGE_READONLY );
+                                          &size, 1 /* ViewShare */, 0, PAGE_READONLY );
             todo_wine_if (test == 5)
             ok( status == (test == 6 ? STATUS_IMAGE_NOT_AT_BASE : STATUS_SUCCESS),
                 "NtMapViewOfSection failed %lx\n", status );
@@ -2397,7 +2419,7 @@
             if (tmp) VirtualFree( tmp, 0, MEM_RELEASE );
             break;
         }
-        DeleteFileA( dll_name );
+        DeleteFileW( dll_name );
         winetest_pop_context();
 #undef DATA_RVA
     }
@@ -2658,51 +2680,18 @@
     return file;
 }
 
-struct ldr_notify_counter
-{
-    WCHAR path[MAX_PATH];
-
-    unsigned int load_count;
-    unsigned int unload_count;
-};
-
-static void CALLBACK ldr_notify_counter_callback(ULONG reason, LDR_DLL_NOTIFICATION_DATA *data, void *context)
-{
-    struct ldr_notify_counter *lnc = context;
-
-    switch (reason)
-    {
-    case LDR_DLL_NOTIFICATION_REASON_LOADED:
-        if (!wcsicmp( data->Loaded.BaseDllName->Buffer, lnc->path ))
-        {
-            lnc->load_count++;
-        }
-        break;
-
-    case LDR_DLL_NOTIFICATION_REASON_UNLOADED:
-        if (!wcsicmp( data->Unloaded.BaseDllName->Buffer, lnc->path ))
-        {
-            lnc->unload_count++;
-        }
-        break;
-    }
-}
-
 static void subtest_export_forwarder_dep_chain( size_t num_chained_export_modules,
                                                 size_t exporter_index,
-                                                BOOL test_static_import,
-                                                ULONG first_module_load_flags )
+                                                BOOL test_static_import )
 {
     size_t num_modules = num_chained_export_modules + !!test_static_import;
     size_t importer_index = test_static_import ? num_modules - 1 : 0;
     DWORD imp_thunk_base_rva, exp_func_base_rva;
     size_t ultimate_depender_index = 0; /* latest module depending on modules earlier in chain */
-    struct ldr_notify_counter lnc;
     char temp_paths[4][MAX_PATH];
     HANDLE temp_files[4];
     UINT_PTR exports[2];
     HMODULE modules[4];
-    void *cookie;
     BOOL res;
     size_t i;
 
@@ -2725,24 +2714,6 @@
                                                    i == importer_index ? &imp_thunk_base_rva : NULL );
     }
 
-    if (first_module_load_flags & DONT_RESOLVE_DLL_REFERENCES)
-    {
-        NTSTATUS status;
-        WCHAR *basename;
-        int cres;
-
-        memset( &lnc, 0, sizeof(lnc) );
-
-        cres = MultiByteToWideChar( CP_ACP, 0, temp_paths[0], -1, lnc.path, ARRAY_SIZE(lnc.path) );
-        ok( cres >= 0, "MultiByteToWideChar returned %d (err %lu)\n", cres, GetLastError() );
-
-        basename = wcsrchr( lnc.path, L'\\' ) + 1;
-        memmove( lnc.path, basename, (char *)lnc.path + sizeof(lnc.path) - (char *)basename );
-
-        status = pLdrRegisterDllNotification( 0, ldr_notify_counter_callback, &lnc, &cookie );
-        ok( !status, "LdrRegisterDllNotification returned %#lx.\n", status );
-    }
-
     if (winetest_debug > 1)
         trace( "Load the entire test DLL chain\n" );
 
@@ -2753,7 +2724,7 @@
         ok( !GetModuleHandleA( temp_paths[i] ), "%s already loaded\n",
             wine_dbgstr_a( temp_paths[i] ) );
 
-        modules[i] = LoadLibraryExA( temp_paths[i], 0, i == 0 ? first_module_load_flags : 0 );
+        modules[i] = LoadLibraryA( temp_paths[i] );
         ok( !!modules[i], "LoadLibraryA(temp_paths[%Iu] = %s) err=%lu\n",
             i, wine_dbgstr_a( temp_paths[i] ), GetLastError() );
 
@@ -2807,16 +2778,6 @@
         ultimate_depender_index = max( ultimate_depender_index, exporter_index );
     }
 
-    if (first_module_load_flags & DONT_RESOLVE_DLL_REFERENCES)
-    {
-        LDR_DATA_TABLE_ENTRY *mod;
-        NTSTATUS status;
-
-        status = LdrFindEntryForAddress( modules[0], &mod );
-        ok( !status, "LdrFindEntryForAddress returned %#lx", status );
-        ok( !(mod->Flags & LDR_PROCESS_ATTACHED), "expected LDR_PROCESS_ATTACHED to be unset (Flags=%#lx)\n", mod->Flags );
-    }
-
     if (winetest_debug > 1)
         trace( "Unreference modules except the ultimate dependant DLL\n" );
 
@@ -2829,6 +2790,7 @@
 
         /* FreeLibrary() should *not* unload the DLL immediately */
         module = GetModuleHandleA( temp_paths[i] );
+        todo_wine_if(i < ultimate_depender_index && i + 1 != importer_index)
         ok( module == modules[i], "modules[%Iu] expected %p, got %p (unloaded?) err=%lu\n",
             i, modules[i], module, GetLastError() );
     }
@@ -2840,6 +2802,7 @@
     {
         HMODULE module = GetModuleHandleA( temp_paths[i] );
 
+        todo_wine_if(i < ultimate_depender_index && i + 1 != importer_index)
         ok( module == modules[i], "modules[%Iu] expected %p, got %p (unloaded?) err=%lu\n",
             i, modules[i], module, GetLastError() );
     }
@@ -2862,17 +2825,6 @@
     for (i = 0; i < num_modules; i++)
     {
         ok( !GetModuleHandleA( temp_paths[i] ), "modules[%Iu] should not be kept loaded (3)\n", i );
-    }
-
-    if (first_module_load_flags & DONT_RESOLVE_DLL_REFERENCES)
-    {
-        NTSTATUS status;
-
-        status = pLdrUnregisterDllNotification( cookie );
-        ok( !status, "LdrUnregisterDllNotification returned %#lx.\n", status );
-
-        ok( lnc.load_count == lnc.unload_count, "got %u/%u for load/unload count of 1st module\n", lnc.load_count, lnc.unload_count );
-        ok( !lnc.load_count || broken(lnc.load_count == 1) /* win7 */, "got %u for load count of first module\n", lnc.load_count );
     }
 
     if (winetest_debug > 1)
@@ -2889,31 +2841,23 @@
 {
     winetest_push_context( "no import" );
     /* export forwarder does not introduce a dependency on its own */
-    subtest_export_forwarder_dep_chain( 2, 0, FALSE, 0 );
+    subtest_export_forwarder_dep_chain( 2, 0, FALSE );
     winetest_pop_context();
 
     winetest_push_context( "static import of export forwarder" );
-    subtest_export_forwarder_dep_chain( 2, 0, TRUE, 0 );
+    subtest_export_forwarder_dep_chain( 2, 0, TRUE );
     winetest_pop_context();
 
     winetest_push_context( "static import of chained export forwarder" );
-    subtest_export_forwarder_dep_chain( 3, 0, TRUE, 0 );
+    subtest_export_forwarder_dep_chain( 3, 0, TRUE );
     winetest_pop_context();
 
     winetest_push_context( "dynamic import of export forwarder" );
-    subtest_export_forwarder_dep_chain( 2, 1, FALSE, 0 );
+    subtest_export_forwarder_dep_chain( 2, 1, FALSE );
     winetest_pop_context();
 
     winetest_push_context( "dynamic import of chained export forwarder" );
-    subtest_export_forwarder_dep_chain( 3, 2, FALSE, 0 );
-    winetest_pop_context();
-
-    winetest_push_context( "static import of dll already loaded with DONT_RESOLVE_DLL_REFERENCES" );
-    subtest_export_forwarder_dep_chain( 1, 0, TRUE, DONT_RESOLVE_DLL_REFERENCES );
-    winetest_pop_context();
-
-    winetest_push_context( "dynamic import of dll already loaded with DONT_RESOLVE_DLL_REFERENCES" );
-    subtest_export_forwarder_dep_chain( 2, 1, FALSE, DONT_RESOLVE_DLL_REFERENCES );
+    subtest_export_forwarder_dep_chain( 3, 2, FALSE );
     winetest_pop_context();
 }
 
@@ -3259,7 +3203,7 @@
         addr = NULL;
         size = 0;
         ret = pNtMapViewOfSection(handle, process, &addr, 0, 0, &offset,
-                                  &size, ViewShare, 0, PAGE_READONLY);
+                                  &size, 1 /* ViewShare */, 0, PAGE_READONLY);
         ok(ret == STATUS_SUCCESS, "NtMapViewOfSection error %#lx\n", ret);
         ret = pNtUnmapViewOfSection(process, addr);
         ok(ret == STATUS_SUCCESS, "NtUnmapViewOfSection error %#lx\n", ret);
@@ -3399,15 +3343,16 @@
     DWORD ret, dummy, i, code, expected_code;
     HANDLE file, thread, process;
     HMODULE hmod;
-    PROCESS_BASIC_INFORMATION pbi;
+    struct PROCESS_BASIC_INFORMATION_PRIVATE pbi;
     DWORD_PTR affinity;
     void *cookie;
 
     trace("phase %d: writing %p at %#lx\n", test_dll_phase, dll_entry_point, target_offset);
 
-    if (pFlsAlloc && !NtCurrentTeb()->Peb->SparePointers[0] /* was FlsCallback */)
-    {
-        fls_list_head = NtCurrentTeb()->FlsSlots->fls_list_entry.Flink;
+    if (pFlsAlloc)
+    {
+        fls_list_head = NtCurrentTeb()->Peb->FlsListHead.Flink ? &NtCurrentTeb()->Peb->FlsListHead
+                : NtCurrentTeb()->FlsSlots->fls_list_entry.Flink;
     }
 
     SetLastError(0xdeadbeef);
@@ -3705,7 +3650,7 @@
 static void test_ExitProcess(void)
 {
 #if defined(__i386__) || defined(__x86_64__) || defined(__aarch64__)
-#pragma pack(push,1)
+#include "pshpack1.h"
 #ifdef __x86_64__
     static struct section_data
     {
@@ -3728,7 +3673,7 @@
         void *target;
     } section_data = { 0x58000040, 0xd61f0000, dll_entry_point };
 #endif
-#pragma pack(pop)
+#include "poppack.h"
     DWORD dummy, file_align;
     HANDLE file, thread, process, hmap, hmap_dup;
     char temp_path[MAX_PATH], dll_name[MAX_PATH], cmdline[MAX_PATH * 2];
@@ -3737,10 +3682,9 @@
     PROCESS_INFORMATION pi;
     STARTUPINFOA si = { sizeof(si) };
     CONTEXT ctx;
-    PROCESS_BASIC_INFORMATION pbi;
+    struct PROCESS_BASIC_INFORMATION_PRIVATE pbi;
     MEMORY_BASIC_INFORMATION mbi;
     DWORD_PTR affinity;
-    PROCESS_PRIORITY_CLASS ppc;
     void *addr;
     LARGE_INTEGER offset;
     SIZE_T size;
@@ -4028,7 +3972,7 @@
     addr = NULL;
     size = 0;
     ret = pNtMapViewOfSection(hmap, pi.hProcess, &addr, 0, 0, &offset,
-                              &size, ViewShare, 0, PAGE_READONLY);
+                              &size, 1 /* ViewShare */, 0, PAGE_READONLY);
     ok(!ret, "NtMapViewOfSection error %#lx\n", ret);
     ret = pNtUnmapViewOfSection(pi.hProcess, addr);
     ok(!ret, "NtUnmapViewOfSection error %#lx\n", ret);
@@ -4080,10 +4024,6 @@
     affinity = 1;
     ret = pNtSetInformationProcess(pi.hProcess, ProcessAffinityMask, &affinity, sizeof(affinity));
     ok(ret == STATUS_PROCESS_IS_TERMINATING, "expected STATUS_PROCESS_IS_TERMINATING, got %#lx\n", ret);
-    ppc.Foreground = FALSE;
-    ppc.PriorityClass = PROCESS_PRIOCLASS_BELOW_NORMAL;
-    ret = pNtSetInformationProcess(pi.hProcess, ProcessPriorityClass, &ppc, sizeof(ppc));
-    ok(ret == STATUS_SUCCESS, "expected STATUS_SUCCESS, got status %#lx\n", ret);
 
     SetLastError(0xdeadbeef);
     ctx.ContextFlags = CONTEXT_INTEGER;
@@ -4171,7 +4111,7 @@
     addr = NULL;
     size = 0;
     ret = pNtMapViewOfSection(hmap, pi.hProcess, &addr, 0, 0, &offset,
-                              &size, ViewShare, 0, PAGE_READONLY);
+                              &size, 1 /* ViewShare */, 0, PAGE_READONLY);
     ok(ret == STATUS_PROCESS_IS_TERMINATING, "expected STATUS_PROCESS_IS_TERMINATING, got %#lx\n", ret);
 
     SetLastError(0xdeadbeef);
@@ -4375,7 +4315,7 @@
     ok(ret, "WriteFile error %ld\n", GetLastError());
 
     /* fill up to delay data */
-    SetFilePointer( hfile, nt_header.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DELAY_IMPORT].VirtualAddress, NULL, FILE_BEGIN );
+    SetFilePointer( hfile, nt_header.OptionalHeader.DataDirectory[IMAGE_DIRECTORY_ENTRY_DELAY_IMPORT].VirtualAddress, NULL, SEEK_SET );
 
     /* delay data */
     idd.Attributes.AllAttributes = 1;
@@ -4396,7 +4336,7 @@
     ok(ret, "WriteFile error %ld\n", GetLastError());
 
     /* fill up to extended delay data */
-    SetFilePointer( hfile, idd.DllNameRVA, NULL, FILE_BEGIN );
+    SetFilePointer( hfile, idd.DllNameRVA, NULL, SEEK_SET );
 
     /* extended delay data */
     SetLastError(0xdeadbeef);
@@ -4411,7 +4351,7 @@
     ret = WriteFile(hfile, test_func, sizeof(test_func), &dummy, NULL);
     ok(ret, "WriteFile error %ld\n", GetLastError());
 
-    SetFilePointer( hfile, idd.ImportAddressTableRVA, NULL, FILE_BEGIN );
+    SetFilePointer( hfile, idd.ImportAddressTableRVA, NULL, SEEK_SET );
 
     for (i = 0; i < ARRAY_SIZE(td); i++)
     {
@@ -4444,7 +4384,7 @@
     ok(ret, "WriteFile error %ld\n", GetLastError());
 
     /* fill up to eof */
-    SetFilePointer( hfile, section.VirtualAddress + section.Misc.VirtualSize, NULL, FILE_BEGIN );
+    SetFilePointer( hfile, section.VirtualAddress + section.Misc.VirtualSize, NULL, SEEK_SET );
     SetEndOfFile( hfile );
     CloseHandle(hfile);
 
@@ -4802,7 +4742,6 @@
     pRtlImageDirectoryEntryToData = (void *)GetProcAddress(ntdll, "RtlImageDirectoryEntryToData");
     pRtlImageNtHeader = (void *)GetProcAddress(ntdll, "RtlImageNtHeader");
     pLdrRegisterDllNotification = (void *)GetProcAddress(ntdll, "LdrRegisterDllNotification");
-    pLdrUnregisterDllNotification = (void *)GetProcAddress(ntdll, "LdrUnregisterDllNotification");
     pFlsAlloc = (void *)GetProcAddress(kernel32, "FlsAlloc");
     pFlsSetValue = (void *)GetProcAddress(kernel32, "FlsSetValue");
     pFlsGetValue = (void *)GetProcAddress(kernel32, "FlsGetValue");
