diff --git a/dlls/comctl32/treeview.c b/dlls/comctl32/treeview.c
--- a/dlls/comctl32/treeview.c
+++ b/dlls/comctl32/treeview.c
@@ -54,6 +54,8 @@
 #include "winnls.h"
 #include "commctrl.h"
 #include "comctl32.h"
+#include "uxtheme.h"
+#include "vssym32.h"
 #include "wine/debug.h"
 #include "wine/exception.h"
 
@@ -195,6 +197,9 @@
 #define GETTXTCOLOR(x)  ((x) == CLR_NONE    ? comctl32_color.clrWindowText : (x))
 #define GETINSCOLOR(x)  ((x) == CLR_DEFAULT ? comctl32_color.clrBtnText    : (x))
 
+static const WCHAR themeClass[] = L"Treeview";
+
+
 typedef VOID (*TREEVIEW_ItemEnumFunc)(TREEVIEW_INFO *, TREEVIEW_ITEM *,LPVOID);
 
 
@@ -517,22 +522,11 @@
     return SendMessageW(infoPtr->hwndNotify, WM_NOTIFY, hdr->idFrom, (LPARAM)hdr);
 }
 
-/*
- * Returns TRUE if the TREEVIEW is still valid after the notify.
- * The notification result is stored in *result if non-NULL.
- */
 static BOOL
-TREEVIEW_SendSimpleNotify(const TREEVIEW_INFO *infoPtr, UINT code, BOOL *result)
+TREEVIEW_SendSimpleNotify(const TREEVIEW_INFO *infoPtr, UINT code)
 {
     NMHDR hdr;
-    HWND hwnd;
-    BOOL notify_val;
-
-    hwnd = infoPtr->hwnd;
-    notify_val = TREEVIEW_SendRealNotify(infoPtr, code, &hdr);
-    if (result)
-        *result = notify_val;
-    return IsWindow(hwnd);
+    return TREEVIEW_SendRealNotify(infoPtr, code, &hdr);
 }
 
 static VOID
@@ -2360,74 +2354,6 @@
     }
 }
 
-/* Draw the (+/-) signs */
-static void TREEVIEW_DrawItemLineSigns(const TREEVIEW_INFO *infoPtr, HDC hdc,
-                                       const TREEVIEW_ITEM *item, LONG centerx, LONG centery,
-                                       COLORREF clrBk)
-{
-    LONG height, width, rectsize, plussize;
-    HPEN new_pen, old_pen;
-
-#if __WINE_COMCTL32_VERSION == 6
-    HTHEME theme = GetWindowTheme(infoPtr->hwnd);
-
-    if (theme)
-    {
-        int state = (item->state & TVIS_EXPANDED) ? GLPS_OPENED : GLPS_CLOSED;
-        RECT glyphRect = item->rect;
-
-        glyphRect.left = item->linesOffset;
-        glyphRect.right = item->stateOffset;
-        DrawThemeBackground(theme, hdc, TVP_GLYPH, state, &glyphRect, NULL);
-        return;
-    }
-#endif
-
-    height = item->rect.bottom - item->rect.top;
-    width = item->stateOffset - item->linesOffset;
-    rectsize = min(height, width) / 4;
-    /* plussize = ceil(rectsize * 3/4) */
-    plussize = (rectsize + 1) * 3 / 4;
-
-    new_pen = CreatePen(PS_SOLID, 0, GETLINECOLOR(infoPtr->clrLine));
-    old_pen = SelectObject(hdc, new_pen);
-
-    Rectangle(hdc, centerx - rectsize - 1, centery - rectsize - 1, centerx + rectsize + 2,
-              centery + rectsize + 2);
-
-    SelectObject(hdc, old_pen);
-    DeleteObject(new_pen);
-
-    /* draw +/- signs with current text color */
-    new_pen = CreatePen(PS_SOLID, 0, GETTXTCOLOR(infoPtr->clrText));
-    old_pen = SelectObject(hdc, new_pen);
-
-    if (height < 18 || width < 18)
-    {
-        MoveToEx(hdc, centerx - plussize + 1, centery, NULL);
-        LineTo(hdc, centerx + plussize, centery);
-
-        if (!(item->state & TVIS_EXPANDED) || (item->state & TVIS_EXPANDPARTIAL))
-        {
-            MoveToEx(hdc, centerx, centery - plussize + 1, NULL);
-            LineTo(hdc, centerx, centery + plussize);
-        }
-    }
-    else
-    {
-        Rectangle(hdc, centerx - plussize + 1, centery - 1, centerx + plussize, centery + 2);
-
-        if (!(item->state & TVIS_EXPANDED) || (item->state & TVIS_EXPANDPARTIAL))
-        {
-            Rectangle(hdc, centerx - 1, centery - plussize + 1, centerx + 2, centery + plussize);
-            SetPixel(hdc, centerx - 1, centery, clrBk);
-            SetPixel(hdc, centerx + 1, centery, clrBk);
-        }
-    }
-
-    SelectObject(hdc, old_pen);
-    DeleteObject(new_pen);
-}
 
 /* Painting *************************************************************/
 
@@ -2508,9 +2434,73 @@
      * Display the (+/-) signs
      */
 
-    if (infoPtr->dwStyle & TVS_HASBUTTONS && item->cChildren)
-        TREEVIEW_DrawItemLineSigns(infoPtr, hdc, item, centerx, centery, clrBk);
-
+    if (infoPtr->dwStyle & TVS_HASBUTTONS)
+    {
+	if (item->cChildren)
+	{
+            HTHEME theme = GetWindowTheme(infoPtr->hwnd);
+            if (theme)
+            {
+                RECT glyphRect = item->rect;
+                glyphRect.left = item->linesOffset;
+                glyphRect.right = item->stateOffset;
+                DrawThemeBackground (theme, hdc, TVP_GLYPH,
+                    (item->state & TVIS_EXPANDED) ? GLPS_OPENED : GLPS_CLOSED,
+                    &glyphRect, NULL);
+            }
+            else
+            {
+                LONG height = item->rect.bottom - item->rect.top;
+                LONG width  = item->stateOffset - item->linesOffset;
+                LONG rectsize = min(height, width) / 4;
+                /* plussize = ceil(rectsize * 3/4) */
+                LONG plussize = (rectsize + 1) * 3 / 4;
+
+                HPEN new_pen  = CreatePen(PS_SOLID, 0, GETLINECOLOR(infoPtr->clrLine));
+                HPEN old_pen  = SelectObject(hdc, new_pen);
+
+                Rectangle(hdc, centerx - rectsize - 1, centery - rectsize - 1,
+                          centerx + rectsize + 2, centery + rectsize + 2);
+
+                SelectObject(hdc, old_pen);
+                DeleteObject(new_pen);
+
+                /* draw +/- signs with current text color */
+                new_pen = CreatePen(PS_SOLID, 0, GETTXTCOLOR(infoPtr->clrText));
+                old_pen = SelectObject(hdc, new_pen);
+
+                if (height < 18 || width < 18)
+                {
+                    MoveToEx(hdc, centerx - plussize + 1, centery, NULL);
+                    LineTo(hdc, centerx + plussize, centery);
+    
+                    if (!(item->state & TVIS_EXPANDED) ||
+                         (item->state & TVIS_EXPANDPARTIAL))
+                    {
+                        MoveToEx(hdc, centerx, centery - plussize + 1, NULL);
+                        LineTo(hdc, centerx, centery + plussize);
+                    }
+                }
+                else
+                {
+                    Rectangle(hdc, centerx - plussize + 1, centery - 1,
+                    centerx + plussize, centery + 2);
+
+                    if (!(item->state & TVIS_EXPANDED) ||
+                         (item->state & TVIS_EXPANDPARTIAL))
+                    {
+                        Rectangle(hdc, centerx - 1, centery - plussize + 1,
+                        centerx + 2, centery + plussize);
+                        SetPixel(hdc, centerx - 1, centery, clrBk);
+                        SetPixel(hdc, centerx + 1, centery, clrBk);
+                    }
+                }
+
+                SelectObject(hdc, old_pen);
+                DeleteObject(new_pen);
+            }
+	}
+    }
     SelectObject(hdc, hbrOld);
     DeleteObject(hbr);
 }
@@ -2902,25 +2892,13 @@
     return 1;
 }
 
-#if __WINE_COMCTL32_VERSION == 6
-static void TREEVIEW_FillThemeBackground(TREEVIEW_INFO *infoPtr, HDC hdc, const RECT *rect)
-{
-    HTHEME theme;
-
-    if (infoPtr->dwStyle & TVS_HASBUTTONS && (theme = GetWindowTheme(infoPtr->hwnd)))
-    {
-        if (IsThemeBackgroundPartiallyTransparent(theme, TVP_GLYPH, 0))
-            FillRect(hdc, rect, (HBRUSH)(COLOR_WINDOW + 1));
-    }
-}
-#endif
-
 static void
 TREEVIEW_Refresh(TREEVIEW_INFO *infoPtr, HDC hdc, const RECT *rc)
 {
     HWND hwnd = infoPtr->hwnd;
     RECT rect = *rc;
     TREEVIEW_ITEM *item;
+    HTHEME theme;
 
     if (infoPtr->clientHeight == 0 || infoPtr->clientWidth == 0)
     {
@@ -2937,9 +2915,11 @@
 	return;
     }
 
-#if __WINE_COMCTL32_VERSION == 6
-    TREEVIEW_FillThemeBackground(infoPtr, hdc, &rect);
-#endif
+    if (infoPtr->dwStyle & TVS_HASBUTTONS && (theme = GetWindowTheme(infoPtr->hwnd)))
+    {
+        if (IsThemeBackgroundPartiallyTransparent(theme, TVP_GLYPH, 0))
+            FillRect(hdc, &rect, (HBRUSH)(COLOR_WINDOW + 1));
+    }
 
     for (item = infoPtr->root->firstChild;
          item != NULL;
@@ -2979,7 +2959,6 @@
         InvalidateRect(infoPtr->hwnd, NULL, TRUE);
 }
 
-#if __WINE_COMCTL32_VERSION == 6
 static BOOL TREEVIEW_InitThemedCheckboxes(TREEVIEW_INFO *info)
 {
     HBITMAP bitmap, old_bitmap;
@@ -3025,7 +3004,6 @@
     info->stateImageHeight = size.cy;
     return TRUE;
 }
-#endif /* __WINE_COMCTL32_VERSION == 6 */
 
 static void
 TREEVIEW_InitCheckboxes(TREEVIEW_INFO *infoPtr)
@@ -3035,10 +3013,8 @@
     HBITMAP hbm, hbmOld;
     HDC hdc, hdcScreen;
 
-#if __WINE_COMCTL32_VERSION == 6
     if (TREEVIEW_InitThemedCheckboxes(infoPtr))
         return;
-#endif
 
     width = 12 * GetDpiForWindow(infoPtr->hwnd) / 96 + 1;
     height = width;
@@ -4217,7 +4193,6 @@
 {
     TREEVIEW_ITEM *item;
     TVHITTESTINFO hit;
-    BOOL notify_result;
 
     TRACE("\n");
     SetFocus(infoPtr->hwnd);
@@ -4236,9 +4211,7 @@
 	return 0;
     TRACE("item %d\n", TREEVIEW_GetItemIndex(infoPtr, item));
 
-    if (TREEVIEW_SendSimpleNotify(infoPtr, NM_DBLCLK, &notify_result) == FALSE)
-        return 0;
-    if (notify_result == FALSE)
+    if (TREEVIEW_SendSimpleNotify(infoPtr, NM_DBLCLK) == FALSE)
     {				/* FIXME! */
 	switch (hit.flags)
 	{
@@ -4285,7 +4258,7 @@
 static LRESULT
 TREEVIEW_LButtonDown(TREEVIEW_INFO *infoPtr, LPARAM lParam)
 {
-    BOOL do_track, do_select, bDoLabelEdit, notify_result;
+    BOOL do_track, do_select, bDoLabelEdit;
     HWND hwnd = infoPtr->hwnd;
     TVHITTESTINFO ht;
 
@@ -4342,13 +4315,8 @@
         (ht.flags & TVHT_ONITEMLABEL) && (infoPtr->selectedItem == ht.hItem);
 
     /* Send NM_CLICK right away */
-    if (!do_track)
-    {
-        if (!TREEVIEW_SendSimpleNotify(infoPtr, NM_CLICK, &notify_result))
-            return 0;
-        if (notify_result)
-            goto setfocus;
-    }
+    if (!do_track && TREEVIEW_SendSimpleNotify(infoPtr, NM_CLICK))
+        goto setfocus;
 
     if (ht.flags & TVHT_ONITEMBUTTON)
     {
@@ -4377,13 +4345,8 @@
         }
     }
 
-    if (do_track)
-    {
-        if (!TREEVIEW_SendSimpleNotify(infoPtr, NM_CLICK, &notify_result))
-            return 0;
-        if (notify_result)
-            goto setfocus;
-    }
+    if (do_track && TREEVIEW_SendSimpleNotify(infoPtr, NM_CLICK))
+        goto setfocus;
 
     if (TREEVIEW_IsFullRowSelect(infoPtr))
         do_select = ht.flags & (TVHT_ONITEMINDENT | TVHT_ONITEMICON | TVHT_ONITEMLABEL | TVHT_ONITEMRIGHT);
@@ -4422,7 +4385,6 @@
 static LRESULT
 TREEVIEW_RButtonDown(TREEVIEW_INFO *infoPtr, LPARAM lParam)
 {
-    BOOL notify_result;
     TVHITTESTINFO ht;
 
     if (infoPtr->hwndEdit)
@@ -4452,9 +4414,7 @@
     else
     {
 	SetFocus(infoPtr->hwnd);
-        if (!TREEVIEW_SendSimpleNotify(infoPtr, NM_RCLICK, &notify_result))
-            return 0;
-        if (!notify_result)
+	if(!TREEVIEW_SendSimpleNotify(infoPtr, NM_RCLICK))
 	{
 	    /* Send a WM_CONTEXTMENU message in response to the RBUTTONUP */
 	    SendMessageW(infoPtr->hwndNotify, WM_CONTEXTMENU,
@@ -5257,8 +5217,8 @@
     /* Make sure actual scrollbar state is consistent with uInternalStatus */
     ShowScrollBar(hwnd, SB_VERT, FALSE);
     ShowScrollBar(hwnd, SB_HORZ, FALSE);
-
-    COMCTL32_OpenThemeForWindow (hwnd, L"Treeview");
+    
+    OpenThemeData (hwnd, themeClass);
 
     return 0;
 }
@@ -5280,7 +5240,7 @@
 	SetWindowLongPtrW(infoPtr->hwndEdit, GWLP_WNDPROC,
 		       (DWORD_PTR)infoPtr->wpEditOrig);
 
-    COMCTL32_CloseThemeForWindow (infoPtr->hwnd);
+    CloseThemeData (GetWindowTheme (infoPtr->hwnd));
 
     /* Deassociate treeview from the window before doing anything drastic. */
     SetWindowLongPtrW(infoPtr->hwnd, 0, 0);
@@ -5380,7 +5340,7 @@
 	break;
 
     case VK_RETURN:
-        TREEVIEW_SendSimpleNotify(infoPtr, NM_RETURN, NULL);
+        TREEVIEW_SendSimpleNotify(infoPtr, NM_RETURN);
         break;
 
     case VK_HOME:
@@ -5523,6 +5483,48 @@
     }
 
     return 0;
+}
+
+/* Draw themed border */
+static BOOL TREEVIEW_NCPaint (const TREEVIEW_INFO *infoPtr, HRGN region, LPARAM lParam)
+{
+    int cxEdge, cyEdge;
+    LONG ex_style;
+    HTHEME theme;
+    HDC dc;
+    RECT r;
+    HRGN cliprgn;
+
+    ex_style = GetWindowLongW(infoPtr->hwnd, GWL_EXSTYLE);
+    if (!(ex_style & WS_EX_CLIENTEDGE))
+        return DefWindowProcW(infoPtr->hwnd, WM_NCPAINT, (WPARAM)region, lParam);
+
+    theme = GetWindowTheme(infoPtr->hwnd);
+    if (!theme)
+        return DefWindowProcW (infoPtr->hwnd, WM_NCPAINT, (WPARAM)region, lParam);
+
+    GetWindowRect(infoPtr->hwnd, &r);
+
+    cxEdge = GetSystemMetrics(SM_CXEDGE);
+    cyEdge = GetSystemMetrics(SM_CYEDGE);
+    cliprgn = CreateRectRgn (r.left + cxEdge, r.top + cyEdge,
+        r.right - cxEdge, r.bottom - cyEdge);
+    if (region != (HRGN)1)
+        CombineRgn (cliprgn, cliprgn, region, RGN_AND);
+    OffsetRect(&r, -r.left, -r.top);
+
+    dc = GetDCEx(infoPtr->hwnd, region, DCX_WINDOW|DCX_INTERSECTRGN);
+
+    if (IsThemeBackgroundPartiallyTransparent (theme, 0, 0))
+        DrawThemeParentBackground(infoPtr->hwnd, dc, &r);
+    DrawThemeBackground (theme, dc, 0, 0, &r, 0);
+    ReleaseDC(infoPtr->hwnd, dc);
+
+    /* Call default proc to get the scrollbars etc. painted */
+    DefWindowProcW (infoPtr->hwnd, WM_NCPAINT, (WPARAM)cliprgn, 0);
+    DeleteObject(cliprgn);
+
+    return TRUE;
 }
 
 static LRESULT
@@ -5664,7 +5666,7 @@
     }
 
     TREEVIEW_Invalidate(infoPtr, infoPtr->selectedItem);
-    TREEVIEW_SendSimpleNotify(infoPtr, NM_SETFOCUS, NULL);
+    TREEVIEW_SendSimpleNotify(infoPtr, NM_SETFOCUS);
     return 0;
 }
 
@@ -5675,9 +5677,20 @@
 
     TREEVIEW_Invalidate(infoPtr, infoPtr->selectedItem);
     UpdateWindow(infoPtr->hwnd);
-    TREEVIEW_SendSimpleNotify(infoPtr, NM_KILLFOCUS, NULL);
+    TREEVIEW_SendSimpleNotify(infoPtr, NM_KILLFOCUS);
     return 0;
 }
+
+/* update theme after a WM_THEMECHANGED message */
+static LRESULT TREEVIEW_ThemeChanged(const TREEVIEW_INFO *infoPtr)
+{
+    HTHEME theme = GetWindowTheme (infoPtr->hwnd);
+    CloseThemeData (theme);
+    OpenThemeData (infoPtr->hwnd, themeClass);
+    InvalidateRect (infoPtr->hwnd, NULL, TRUE);
+    return 0;
+}
+
 
 static LRESULT WINAPI
 TREEVIEW_WindowProc(HWND hwnd, UINT uMsg, WPARAM wParam, LPARAM lParam)
@@ -5846,11 +5859,6 @@
     case WM_GETDLGCODE:
 	return DLGC_WANTARROWS | DLGC_WANTCHARS;
 
-    case WM_GETOBJECT:
-        if ((LONG)lParam == OBJID_QUERYCLASSNAMEIDX)
-            return 0x10019;
-	return DefWindowProcW(hwnd, uMsg, wParam, lParam);
-
     case WM_GETFONT:
 	return TREEVIEW_GetFont(infoPtr);
 
@@ -5884,7 +5892,7 @@
         goto def;
 
     case WM_NCPAINT:
-        return COMCTL32_NCPaint (infoPtr->hwnd, wParam, lParam, NULL);
+        return TREEVIEW_NCPaint (infoPtr, (HRGN)wParam, lParam);
 
     case WM_NOTIFY:
 	return TREEVIEW_Notify(infoPtr, wParam, lParam);
@@ -5927,7 +5935,7 @@
 	return TREEVIEW_HandleTimer(infoPtr, wParam);
 
     case WM_THEMECHANGED:
-        return COMCTL32_ThemeChanged (infoPtr->hwnd, L"Treeview", TRUE, TRUE);
+        return TREEVIEW_ThemeChanged (infoPtr);
 
     case WM_VSCROLL:
 	return TREEVIEW_VScroll(infoPtr, wParam);
@@ -5973,6 +5981,14 @@
     RegisterClassW(&wndClass);
 }
 
+
+VOID
+TREEVIEW_Unregister(void)
+{
+    UnregisterClassW(WC_TREEVIEWW, NULL);
+}
+
+
 /* Tree Verification ****************************************************/
 
 static inline void
