diff --git a/dlls/propsys/propsys_main.c b/dlls/propsys/propsys_main.c
--- a/dlls/propsys/propsys_main.c
+++ b/dlls/propsys/propsys_main.c
@@ -28,7 +28,6 @@
 #include "objbase.h"
 #include "rpcproxy.h"
 #include "propsys.h"
-#include "propkey.h"
 #include "wine/debug.h"
 
 #include "propsys_private.h"
@@ -93,40 +92,12 @@
 
 static IClassFactory InMemoryPropertyStoreFactory = { &InMemoryPropertyStoreFactoryVtbl };
 
-static IPropertySystem propsys;
-
-static HRESULT WINAPI propsys_factory_CreateInstance(IClassFactory *iface, IUnknown *outer, REFIID iid, void **out)
-{
-    TRACE("%p, %p, %s, %p\n", iface, outer, debugstr_guid(iid), out);
-
-    *out = NULL;
-    if (outer)
-        return CLASS_E_NOAGGREGATION;
-    return IPropertySystem_QueryInterface(&propsys, iid, out);
-}
-
-static const IClassFactoryVtbl PropertySystemFactoryVtbl =
-{
-    ClassFactory_QueryInterface,
-    ClassFactory_AddRef,
-    ClassFactory_Release,
-    propsys_factory_CreateInstance,
-    ClassFactory_LockServer
-};
-
-static IClassFactory PropertySystemFactory = { &PropertySystemFactoryVtbl };
-
 HRESULT WINAPI DllGetClassObject(REFCLSID rclsid, REFIID riid, LPVOID *ppv)
 {
     if(IsEqualGUID(&CLSID_InMemoryPropertyStore, rclsid)) {
         TRACE("(CLSID_InMemoryPropertyStore %s %p)\n", debugstr_guid(riid), ppv);
         return IClassFactory_QueryInterface(&InMemoryPropertyStoreFactory, riid, ppv);
     }
-    if (IsEqualGUID(&CLSID_PropertySystem, rclsid))
-    {
-        TRACE("(CLSID_PropertySystem %s %p)\n", debugstr_guid(riid), ppv);
-        return IClassFactory_QueryInterface(&PropertySystemFactory, riid, ppv);
-    }
 
     FIXME("%s %s %p\n", debugstr_guid(rclsid), debugstr_guid(riid), ppv);
     return CLASS_E_CLASSNOTAVAILABLE;
@@ -134,7 +105,6 @@
 
 static HRESULT WINAPI propsys_QueryInterface(IPropertySystem *iface, REFIID riid, void **obj)
 {
-    TRACE("(%p, %s, %p)\n", iface, debugstr_guid(riid), obj);
     *obj = NULL;
 
     if (IsEqualIID(riid, &IID_IPropertySystem) || IsEqualIID(riid, &IID_IUnknown)) {
@@ -157,46 +127,17 @@
     return 1;
 }
 
-static HRESULT propdesc_get_by_name( const WCHAR *name, IPropertyDescription **desc );
-static HRESULT propdesc_get_by_key( const PROPERTYKEY *key, IPropertyDescription **desc );
-
 static HRESULT WINAPI propsys_GetPropertyDescription(IPropertySystem *iface,
     REFPROPERTYKEY propkey, REFIID riid, void **ppv)
 {
-    HRESULT hr;
-    IPropertyDescription *desc;
-
-    FIXME("(%p, %s, %s, %p): semi-stub!\n", iface, debugstr_propkey(propkey), debugstr_guid(riid), ppv);
-
-    if (!ppv)
-        return E_INVALIDARG;
-    *ppv = NULL;
-    hr = propdesc_get_by_key( propkey, &desc );
-    if (FAILED( hr ))
-        return hr;
-
-    hr = IPropertyDescription_QueryInterface( desc, riid, ppv );
-    IPropertyDescription_Release( desc );
-    return hr;
+    return PSGetPropertyDescription(propkey, riid, ppv);
 }
 
 static HRESULT WINAPI propsys_GetPropertyDescriptionByName(IPropertySystem *iface,
     LPCWSTR canonical_name, REFIID riid, void **ppv)
 {
-    HRESULT hr;
-    IPropertyDescription *desc;
-
-    FIXME("(%p, %s, %s, %p): stub\n", iface, debugstr_w(canonical_name), debugstr_guid(riid), ppv);
-
-    if (!ppv)
-        return E_INVALIDARG;
-    *ppv = NULL;
-    hr = propdesc_get_by_name( canonical_name, &desc );
-    if (FAILED( hr ))
-        return hr;
-    hr = IPropertyDescription_QueryInterface( desc, riid, ppv );
-    IPropertyDescription_Release( desc );
-    return hr;
+    FIXME("%s %s %p: stub\n", debugstr_w(canonical_name), debugstr_guid(riid), ppv);
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI propsys_GetPropertyDescriptionListFromString(IPropertySystem *iface,
@@ -275,22 +216,14 @@
 HRESULT WINAPI PSUnregisterPropertySchema(PCWSTR path)
 {
     FIXME("%s stub\n", debugstr_w(path));
+
     return E_NOTIMPL;
 }
 
 HRESULT WINAPI PSGetPropertyDescription(REFPROPERTYKEY propkey, REFIID riid, void **ppv)
 {
-    HRESULT hr;
-    IPropertySystem *system;
-
-    TRACE("%p, %p, %p\n", propkey, riid, ppv);
-    hr = IPropertySystem_QueryInterface(&propsys, &IID_IPropertySystem, (void **)&system);
-    if (SUCCEEDED(hr))
-    {
-        hr = IPropertySystem_GetPropertyDescription(system, propkey, riid, ppv);
-        IPropertySystem_Release(system);
-    }
-    return hr;
+    FIXME("%p, %p, %p\n", propkey, riid, ppv);
+    return E_NOTIMPL;
 }
 
 HRESULT WINAPI PSGetPropertyDescriptionListFromString(LPCWSTR proplist, REFIID riid, void **ppv)
@@ -299,49 +232,10 @@
     return E_NOTIMPL;
 }
 
-HRESULT WINAPI PSGetNameFromPropertyKey(REFPROPERTYKEY key, LPWSTR *name)
-{
-    HRESULT hr;
-    IPropertySystem *system;
-
-    TRACE("(%s, %p)\n", debugstr_propkey(key), name);
-    hr = IPropertySystem_QueryInterface(&propsys, &IID_IPropertySystem, (void **)&system);
-    if (SUCCEEDED(hr))
-    {
-        IPropertyDescription *desc;
-
-        hr = IPropertySystem_GetPropertyDescription(system, key, &IID_IPropertyDescription, (void **)&desc);
-        if (SUCCEEDED(hr))
-        {
-            hr = IPropertyDescription_GetCanonicalName(desc, name);
-            IPropertyDescription_Release(desc);
-        }
-        IPropertySystem_Release(system);
-    }
-
-    return hr;
-}
-
 HRESULT WINAPI PSGetPropertyKeyFromName(PCWSTR name, PROPERTYKEY *key)
 {
-    HRESULT hr;
-    IPropertySystem *system;
-
-    TRACE("%s, %p\n", debugstr_w(name), debugstr_propkey(key));
-    hr = IPropertySystem_QueryInterface(&propsys, &IID_IPropertySystem, (void **)&system);
-    if (SUCCEEDED(hr))
-    {
-        IPropertyDescription *desc;
-        hr = IPropertySystem_GetPropertyDescriptionByName(system, name, &IID_IPropertyDescription, (void **)&desc);
-        if (SUCCEEDED(hr))
-        {
-            hr = IPropertyDescription_GetPropertyKey(desc, key);
-            IPropertyDescription_Release(desc);
-        }
-        IPropertySystem_Release(system);
-    }
-
-    return hr;
+    FIXME("%s, %p\n", debugstr_w(name), key);
+    return E_NOTIMPL;
 }
 
 HRESULT WINAPI PSRefreshPropertySchema(void)
@@ -580,330 +474,3 @@
 
     return PropertyStore_CreateInstance(NULL, riid, ppv);
 }
-
-struct property_description
-{
-    IPropertyDescription IPropertyDescription_iface;
-    LONG ref;
-
-    PROPERTYKEY key;
-    VARTYPE type;
-    PROPDESC_TYPE_FLAGS type_flags;
-    WCHAR *canonical_name;
-};
-
-static inline struct property_description *
-impl_from_IPropertyDescription( IPropertyDescription *iface )
-{
-    return CONTAINING_RECORD( iface, struct property_description, IPropertyDescription_iface );
-}
-
-static HRESULT WINAPI propdesc_QueryInterface( IPropertyDescription *iface, REFIID iid, void **out )
-{
-    TRACE( "(%p, %s, %p)\n", iface, debugstr_guid( iid ), out );
-    *out = NULL;
-
-    if (IsEqualGUID( &IID_IUnknown, iid ) ||
-        IsEqualGUID( &IID_IPropertyDescription, iid ))
-    {
-        *out = iface;
-        IUnknown_AddRef( iface );
-        return S_OK;
-    }
-
-    FIXME( "%s not implemented\n", debugstr_guid( iid ) );
-    return E_NOINTERFACE;
-}
-
-static ULONG WINAPI propdesc_AddRef( IPropertyDescription *iface )
-{
-    struct property_description *propdesc = impl_from_IPropertyDescription( iface );
-    TRACE( "(%p)\n", iface );
-    return InterlockedIncrement( &propdesc->ref );
-}
-
-static ULONG WINAPI propdesc_Release( IPropertyDescription *iface )
-{
-    struct property_description *propdesc = impl_from_IPropertyDescription( iface );
-    ULONG ref;
-
-    TRACE( "(%p)\n", iface );
-    ref = InterlockedDecrement( &propdesc->ref );
-    if (!ref)
-    {
-        free( propdesc->canonical_name );
-        free( propdesc );
-    }
-
-    return ref;
-}
-
-static HRESULT WINAPI propdesc_GetPropertyKey( IPropertyDescription *iface, PROPERTYKEY *pkey )
-{
-    struct property_description *propdesc = impl_from_IPropertyDescription( iface );
-
-    TRACE( "(%p, %p)\n", iface, pkey );
-    *pkey = propdesc->key;
-    return S_OK;
-}
-
-static HRESULT WINAPI propdesc_GetCanonicalName( IPropertyDescription *iface, LPWSTR *name )
-{
-    struct property_description *propdesc = impl_from_IPropertyDescription( iface );
-
-    TRACE( "(%p, %p)\n", iface, name );
-    *name = CoTaskMemAlloc( (wcslen( propdesc->canonical_name ) + 1) * sizeof( WCHAR ) );
-    if (!*name)
-        return E_OUTOFMEMORY;
-    wcscpy( *name, propdesc->canonical_name );
-    return S_OK;
-}
-
-static HRESULT WINAPI propdesc_GetPropertyType( IPropertyDescription *iface, VARTYPE *vt )
-{
-    struct property_description *propdesc = impl_from_IPropertyDescription( iface );
-
-    TRACE( "(%p, %p)\n", iface, vt );
-    *vt = propdesc->type;
-    return S_OK;
-}
-
-static HRESULT WINAPI propdesc_GetDisplayName( IPropertyDescription *iface, LPWSTR *name )
-{
-    FIXME( "(%p, %p) semi-stub!\n", iface, name );
-    return IPropertyDescription_GetCanonicalName( iface, name );
-}
-
-static HRESULT WINAPI propdesc_GetEditInvitation( IPropertyDescription *iface, LPWSTR *invite )
-{
-    FIXME( "(%p, %p) stub!\n", iface, invite );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetTypeFlags( IPropertyDescription *iface, PROPDESC_TYPE_FLAGS mask,
-                                             PROPDESC_TYPE_FLAGS *flags )
-{
-    struct property_description *propdesc = impl_from_IPropertyDescription( iface );
-
-    TRACE( "(%p, %#x, %p)\n", iface, mask, flags );
-    *flags = mask & propdesc->type_flags;
-    return S_OK;
-}
-
-static HRESULT WINAPI propdesc_GetViewFlags( IPropertyDescription *iface, PROPDESC_VIEW_FLAGS *flags )
-{
-    FIXME( "(%p, %p) stub!\n", iface, flags );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetDefaultColumnWidth( IPropertyDescription *iface, UINT *chars )
-{
-    FIXME( "(%p, %p) stub!\n", iface, chars );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetDisplayType( IPropertyDescription *iface, PROPDESC_DISPLAYTYPE *type )
-{
-    FIXME( "(%p, %p) stub!\n", iface, type );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetColumnState( IPropertyDescription *iface, SHCOLSTATEF *flags )
-{
-    FIXME( "(%p, %p) stub!\n", iface, flags );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetGroupingRange( IPropertyDescription *iface, PROPDESC_GROUPING_RANGE *range )
-{
-    FIXME( "(%p, %p) stub!\n", iface, range );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetRelativeDescriptionType( IPropertyDescription *iface,
-                                                           PROPDESC_RELATIVEDESCRIPTION_TYPE *type )
-{
-    FIXME( "(%p, %p) stub!\n", iface, type );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetRelativeDescription( IPropertyDescription *iface, REFPROPVARIANT propvar1,
-                                                       REFPROPVARIANT propvar2, LPWSTR *desc1, LPWSTR *desc2 )
-{
-    FIXME( "(%p, %p, %p, %p, %p) stub!\n", iface, propvar1, propvar2, desc1, desc2 );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetSortDescription( IPropertyDescription *iface, PROPDESC_SORTDESCRIPTION *psd )
-{
-    FIXME( "(%p, %p) stub!\n", iface, psd );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetSortDescriptionLabel( IPropertyDescription *iface, BOOL descending, LPWSTR *desc )
-{
-    FIXME( "(%p, %d, %p) stub!\n", iface, descending, desc );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetAggregationType( IPropertyDescription *iface, PROPDESC_AGGREGATION_TYPE *type )
-{
-    FIXME( "(%p, %p) stub!\n", iface, type );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetConditionType( IPropertyDescription *iface, PROPDESC_CONDITION_TYPE *cond_type,
-                                                 CONDITION_OPERATION *op_default )
-{
-    FIXME( "(%p, %p, %p) stub!\n", iface, cond_type, op_default );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_GetEnumTypeList( IPropertyDescription *iface, REFIID riid, void **out )
-{
-    FIXME( "(%p, %s, %p) stub!\n", iface, debugstr_guid( riid ), out );
-    *out = NULL;
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_CoerceToCanonicalValue( IPropertyDescription *iface, PROPVARIANT *propvar )
-{
-    FIXME( "(%p, %p) stub!\n", iface, propvar );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_FormatForDisplay( IPropertyDescription *iface, REFPROPVARIANT propvar,
-                                        PROPDESC_FORMAT_FLAGS flags, LPWSTR *display )
-{
-    FIXME( "(%p, %p, %#x, %p) stub!\n", iface, propvar, flags, display );
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI propdesc_IsValueCanonical( IPropertyDescription *iface, REFPROPVARIANT propvar )
-{
-    FIXME( "(%p, %p) stub!\n", iface, propvar );
-    return E_NOTIMPL;
-}
-
-const static IPropertyDescriptionVtbl property_description_vtbl =
-{
-    /* IUnknown */
-    propdesc_QueryInterface,
-    propdesc_AddRef,
-    propdesc_Release,
-    /* IPropertyDescription */
-    propdesc_GetPropertyKey,
-    propdesc_GetCanonicalName,
-    propdesc_GetPropertyType,
-    propdesc_GetDisplayName,
-    propdesc_GetEditInvitation,
-    propdesc_GetTypeFlags,
-    propdesc_GetViewFlags,
-    propdesc_GetDefaultColumnWidth,
-    propdesc_GetDisplayType,
-    propdesc_GetColumnState,
-    propdesc_GetGroupingRange,
-    propdesc_GetRelativeDescriptionType,
-    propdesc_GetRelativeDescription,
-    propdesc_GetSortDescription,
-    propdesc_GetSortDescriptionLabel,
-    propdesc_GetAggregationType,
-    propdesc_GetConditionType,
-    propdesc_GetEnumTypeList,
-    propdesc_CoerceToCanonicalValue,
-    propdesc_FormatForDisplay,
-    propdesc_IsValueCanonical
-};
-
-struct system_property_description
-{
-    const WCHAR *canonical_name;
-    const PROPERTYKEY *key;
-
-    VARTYPE type;
-};
-
-/* It may be ideal to construct rb_trees for looking up property descriptions by name and key if this array gets large
- * enough. */
-static struct system_property_description system_properties[] =
-{
-    {L"System.Devices.ContainerId", &PKEY_Devices_ContainerId, VT_CLSID},
-    {L"System.Devices.InterfaceClassGuid", &PKEY_Devices_InterfaceClassGuid, VT_CLSID},
-    {L"System.Devices.DeviceInstanceId", &PKEY_Devices_DeviceInstanceId, VT_LPWSTR},
-    {L"System.Devices.InterfaceEnabled", &PKEY_Devices_InterfaceEnabled, VT_BOOL},
-    {L"System.Devices.ClassGuid", &PKEY_Devices_ClassGuid, VT_CLSID},
-    {L"System.Devices.CompatibleIds", &PKEY_Devices_CompatibleIds, VT_VECTOR | VT_LPWSTR},
-    {L"System.Devices.DeviceCapabilities", &PKEY_Devices_DeviceCapabilities, VT_UI2},
-    {L"System.Devices.DeviceHasProblem", &PKEY_Devices_DeviceHasProblem, VT_BOOL},
-    {L"System.Devices.DeviceManufacturer", &PKEY_Devices_DeviceManufacturer, VT_LPWSTR},
-    {L"System.Devices.HardwareIds", &PKEY_Devices_HardwareIds, VT_VECTOR | VT_LPWSTR},
-    {L"System.Devices.Parent", &PKEY_Devices_Parent, VT_LPWSTR},
-    {L"System.ItemNameDisplay", &PKEY_ItemNameDisplay, VT_LPWSTR},
-    {L"System.Devices.Category", &PKEY_Devices_Category, VT_VECTOR | VT_LPWSTR},
-    {L"System.Devices.CategoryIds", &PKEY_Devices_CategoryIds, VT_VECTOR | VT_LPWSTR},
-    {L"System.Devices.CategoryPlural", &PKEY_Devices_CategoryPlural, VT_VECTOR | VT_LPWSTR},
-    {L"System.Devices.Connected", &PKEY_Devices_Connected, VT_BOOL},
-    {L"System.Devices.GlyphIcon", &PKEY_Devices_GlyphIcon, VT_LPWSTR},
-    {L"System.Devices.DevObjectType", &PKEY_Devices_DevObjectType, VT_UI4},
-    {L"System.DeviceInterface.Bluetooth.DeviceAddress", &PKEY_DeviceInterface_Bluetooth_DeviceAddress, VT_LPWSTR},
-    {L"System.Devices.Aep.AepId", &PKEY_Devices_Aep_AepId, VT_LPWSTR},
-    {L"System.Devices.Aep.ProtocolId", &PKEY_Devices_Aep_ProtocolId, VT_CLSID},
-    {L"System.Devices.Aep.IsPaired", &PKEY_Devices_Aep_IsPaired, VT_BOOL},
-    {L"System.Devices.Aep.IsConnected", &PKEY_Devices_Aep_IsConnected, VT_BOOL},
-    {L"System.Devices.Aep.IsConnected", &PKEY_Devices_Aep_IsConnected, VT_BOOL},
-    {L"System.Devices.Aep.Bluetooth.Le.AddressType", &PKEY_Devices_Aep_Bluetooth_Le_AddressType, VT_UI1},
-    {L"System.Devices.Aep.DeviceAddress", &PKEY_Devices_Aep_DeviceAddress, VT_LPWSTR},
-    {L"System.Devices.AepContainer.IsPaired", &PKEY_Devices_AepContainer_IsPaired, VT_BOOL},
-    {L"System.Devices.AepService.ProtocolId", &PKEY_Devices_AepService_ProtocolId, VT_CLSID},
-};
-
-static HRESULT propdesc_from_system_property( const struct system_property_description *desc, IPropertyDescription **out )
-{
-    struct property_description *propdesc;
-
-    propdesc = calloc( 1, sizeof( *propdesc ) );
-    if (!propdesc)
-        return E_OUTOFMEMORY;
-
-    propdesc->IPropertyDescription_iface.lpVtbl = &property_description_vtbl;
-    if (!(propdesc->canonical_name = wcsdup( desc->canonical_name )))
-    {
-        free( propdesc );
-        return E_OUTOFMEMORY;
-    }
-    propdesc->key = *desc->key;
-    propdesc->type = desc->type;
-    propdesc->type_flags = PDTF_ISINNATE;
-    if (propdesc->type & VT_VECTOR)
-        propdesc->type_flags |= PDTF_MULTIPLEVALUES;
-    propdesc->ref = 1;
-
-    *out = &propdesc->IPropertyDescription_iface;
-    return S_OK;
-}
-
-static HRESULT propdesc_get_by_name( const WCHAR *name, IPropertyDescription **desc )
-{
-    SIZE_T i;
-
-    for (i = 0; i < ARRAY_SIZE( system_properties ); i++)
-    {
-        if (!wcscmp( name, system_properties[i].canonical_name ))
-            return propdesc_from_system_property( &system_properties[i], desc );
-    }
-
-    return TYPE_E_ELEMENTNOTFOUND;
-}
-
-static HRESULT propdesc_get_by_key( const PROPERTYKEY *key, IPropertyDescription **desc )
-{
-    SIZE_T i;
-
-    for (i = 0; i < ARRAY_SIZE( system_properties ); i++)
-    {
-        if (!memcmp( key, system_properties[i].key, sizeof( *key ) ))
-            return propdesc_from_system_property( &system_properties[i], desc );
-    }
-
-    return TYPE_E_ELEMENTNOTFOUND;
-}
