diff --git a/dlls/windowscodecs/tests/converter.c b/dlls/windowscodecs/tests/converter.c
--- a/dlls/windowscodecs/tests/converter.c
+++ b/dlls/windowscodecs/tests/converter.c
@@ -233,7 +233,7 @@
 
 static BOOL compare_bits(const struct bitmap_data *expect, UINT buffersize, const BYTE *converted_bits)
 {
-    BOOL equal, is_float = FALSE;
+    BOOL equal;
 
     if (IsEqualGUID(expect->format, &GUID_WICPixelFormat32bppBGR))
     {
@@ -248,14 +248,11 @@
                 break;
             }
     }
-    else if (IsEqualGUID(expect->format, &GUID_WICPixelFormat32bppGrayFloat)
-            || IsEqualGUID(expect->format, &GUID_WICPixelFormat128bppRGBFloat)
-            || IsEqualGUID(expect->format, &GUID_WICPixelFormat128bppRGBAFloat))
+    else if (IsEqualGUID(expect->format, &GUID_WICPixelFormat32bppGrayFloat))
     {
         UINT i;
         const float *a=(const float*)expect->bits, *b=(const float*)converted_bits;
         equal=TRUE;
-        is_float = TRUE;
         for (i=0; i<(buffersize/4); i++)
             if (!near_equal(a[i], b[i]))
             {
@@ -300,28 +297,14 @@
     if (!equal && winetest_debug > 1)
     {
         UINT i, bps;
-
         bps = expect->bpp / 8;
         if (!bps) bps = buffersize;
         printf("converted_bits (%u bytes):\n    ", buffersize);
-        if (is_float)
+        for (i = 0; i < buffersize; i++)
         {
-            const float *src = (const float *)converted_bits;
-            for (i = 0; i < buffersize / 4; i++)
-            {
-                printf("%f,", src[i]);
-                if (!((i + 1) % 32)) printf("\n    ");
-                else if (!((i+1) % bps)) printf(" ");
-            }
-        }
-        else
-        {
-            for (i = 0; i < buffersize; i++)
-            {
-                printf("%u,", converted_bits[i]);
-                if (!((i + 1) % 32)) printf("\n    ");
-                else if (!((i+1) % bps)) printf(" ");
-            }
+            printf("%u,", converted_bits[i]);
+            if (!((i + 1) % 32)) printf("\n    ");
+            else if (!((i+1) % bps)) printf(" ");
         }
         printf("\n");
     }
@@ -449,19 +432,6 @@
     0,255,255, 255,0,255, 255,255,0, 255,255,255, 0,255,255, 255,0,255, 255,255,0, 255,255,255};
 static const struct bitmap_data testdata_24bppBGR = {
     &GUID_WICPixelFormat24bppBGR, 24, bits_24bppBGR, 32, 2, 96.0, 96.0};
-
-static const BYTE bits_24bppBGR_BW[] = {
-    0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255,
-    0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255,
-    0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255,
-    0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255,
-
-    255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0,
-    255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0,
-    255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0,
-    255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0, 255,255,255, 0,0,0};
-static const struct bitmap_data testdata_24bppBGR_BW = {
-    &GUID_WICPixelFormat24bppBGR, 24, bits_24bppBGR_BW, 32, 2, 96.0, 96.0};
 
 static const BYTE bits_24bppRGB[] = {
     0,0,255, 0,255,0, 255,0,0, 0,0,0, 0,0,255, 0,255,0, 255,0,0, 0,0,0,
@@ -666,84 +636,6 @@
 static const struct bitmap_data testdata_64bppRGBA_2 = {
     &GUID_WICPixelFormat64bppRGBA, 64, (BYTE*)bits_64bppRGBA_2, 3, 2, 96.0, 96.0};
 
-static const float bits_96bppRGBFloat[] = {
-    0.0f,0.0f,0.0f, 0.0f,1.0f,0.0f, 0.214039f,0.214053f,0.214039f,
-    1.0f,1.0f,1.0f, 0.000012f,0.000012f,0.000012f, 0.0f,0.0f,0.000012f};
-static const struct bitmap_data testdata_96bppRGBFloat = {
-    &GUID_WICPixelFormat96bppRGBFloat, 96, (const BYTE *)bits_96bppRGBFloat, 3, 2, 96.0, 96.0};
-
-static const float bits_96bppRGBFloat_2[] = {
-    0.0f,0.0f,0.0f, 0.0f,1.0f,0.0f, 1.0f,0.0f,0.0f,
-    0.0f,0.0f,1.0f, 0.0f,0.205079f,0.0f, 0.205079f,0.0f,0.0f};
-static const struct bitmap_data testdata_96bppRGBFloat_2 = {
-    &GUID_WICPixelFormat96bppRGBFloat, 96, (const BYTE *)bits_96bppRGBFloat_2, 3, 2, 96.0, 96.0};
-
-static const BYTE bits_32bppBGRA_3[] = {
-    0,0,0,255, 0,255,0,255, 0,0,255,255,
-    255,0,0,255, 0,125,0,255, 0,0,125,255};
-static const struct bitmap_data testdata_32bppBGRA_3 = {
-    &GUID_WICPixelFormat32bppBGRA, 32, bits_32bppBGRA_3, 3, 2, 96.0, 96.0};
-
-static const BYTE bits_32bppBGRA_4[] = {
-    0,0,0,255, 255,255,255,255, 125,125,125,255,
-    255,255,255,255, 125,125,125,255, 0,0,0,255};
-static const struct bitmap_data testdata_32bppBGRA_4 = {
-    &GUID_WICPixelFormat32bppBGRA, 32, bits_32bppBGRA_4, 3, 2, 96.0, 96.0};
-
-static const WORD bits_48bppRGBHalf[] = {
-    0,0,0, 0,0x3c00,0, 0x3c00,0,0,
-    0,0,0x3c00, 0,0x3290,0, 0x3290,0,0 };
-static const struct bitmap_data testdata_48bppRGBHalf = {
-    &GUID_WICPixelFormat48bppRGBHalf, 48, (const BYTE *)bits_48bppRGBHalf, 3, 2, 96.0, 96.0};
-
-static const WORD bits_16bppGrayHalf[] = {
-    0, 0x3c00, 0x3290,
-    0x3c00, 0x3290, 0 };
-static const struct bitmap_data testdata_16bppGrayHalf = {
-    &GUID_WICPixelFormat16bppGrayHalf, 16, (const BYTE *)bits_16bppGrayHalf, 3, 2, 96.0, 96.0};
-
-static const float bits_128bppRGBFloat[] = {
-    0.0f,0.0f,0.0f,1.0f, 0.0f,1.0f,0.0f,1.0f, 0.214039f,0.214053f,0.214039f,1.0f,
-    1.0f,1.0f,1.0f,1.0f, 0.000012f,0.000012f,0.000012f,1.0f, 0.0f,0.0f,0.000012f,1.0f,};
-static const struct bitmap_data testdata_128bppRGBFloat = {
-    &GUID_WICPixelFormat128bppRGBFloat, 128, (const BYTE *)bits_128bppRGBFloat, 3, 2, 96.0, 96.0};
-
-static const float bits_128bppRGBFloat_2[] = {
-    0.0f,0.0f,0.0f,1.0f, 0.0f,1.0f,0.0f,1.0f, 1.0f,0.0f,0.0f,1.0f,
-    0.0f,0.0f,1.0f,1.0f, 0.0f,0.205079f,0.0f,1.0f, 0.205079f,0.0f,0.0f,1.0f};
-static const struct bitmap_data testdata_128bppRGBFloat_2 = {
-    &GUID_WICPixelFormat128bppRGBFloat, 128, (const BYTE *)bits_128bppRGBFloat_2, 3, 2, 96.0, 96.0};
-
-static const float bits_128bppRGBFloat_3[] = {
-    0.0f,0.0f,0.0f,1.0f, 1.0f,1.0f,1.0f,1.0f, 0.205079f,0.205079f,0.205079f,1.0f,
-    1.0f,1.0f,1.0f,1.0f, 0.205079f,0.205079f,0.205079f,1.0f, 0.0f,0.0f,0.0f,1.0f};
-static const struct bitmap_data testdata_128bppRGBFloat_3 = {
-    &GUID_WICPixelFormat128bppRGBFloat, 128, (const BYTE *)bits_128bppRGBFloat_3, 3, 2, 96.0, 96.0};
-
-static const BYTE bits_24bppBGR_2[] = {
-    0,0,0, 0,255,0, 0,0,255,
-    255,0,0, 0,125,0, 0,0,125};
-static const struct bitmap_data testdata_24bppBGR_2 = {
-    &GUID_WICPixelFormat24bppBGR, 24, bits_24bppBGR_2, 3, 2, 96.0, 96.0};
-
-static const BYTE bits_32bppBGRA_2[] = {
-    0,0,0,0, 0,255,0,128, 0,0,255,255,
-    255,0,0,0, 0,125,0,255, 0,0,125,128};
-static const struct bitmap_data testdata_32bppBGRA_2 = {
-    &GUID_WICPixelFormat32bppBGRA, 32, bits_32bppBGRA_2, 3, 2, 96.0, 96.0};
-
-static const float bits_128bppRGBAFloat[] = {
-    0.0f,0.0f,0.0f,1.0f, 0.0f,1.0f,0.0f,1.0f, 1.0f,0.0f,0.0f,1.0f,
-    0.0f,0.0f,1.0f,1.0f, 0.0f,0.205079f,0.0f,1.0f, 0.205079f,0.0f,0.0f,1.0f};
-static const struct bitmap_data testdata_128bppRGBAFloat = {
-    &GUID_WICPixelFormat128bppRGBAFloat, 128, (const BYTE *)bits_128bppRGBAFloat, 3, 2, 96.0, 96.0};
-
-static const float bits_128bppRGBAFloat_2[] = {
-    0.0f,0.0f,0.0f,0.0f, 0.0f,1.0f,0.0f,0.501961f, 1.0f,0.0f,0.0f,1.0f,
-    0.0f,0.0f,1.0f,0.0f, 0.0f,0.205079f,0.0f,1.0f, 0.205079f,0.0f,0.0f,0.5019f};
-static const struct bitmap_data testdata_128bppRGBAFloat_2 = {
-    &GUID_WICPixelFormat128bppRGBAFloat, 128, (const BYTE *)bits_128bppRGBAFloat_2, 3, 2, 96.0, 96.0};
-
 static void test_conversion(const struct bitmap_data *src, const struct bitmap_data *dst, const char *name, BOOL todo)
 {
     BitmapTestSrc *src_obj;
@@ -815,178 +707,6 @@
     DeleteTestBitmap(src_obj);
 }
 
-static void test_can_convert(void)
-{
-#define WIC_PIXEL_FORMAT(fmt) #fmt, &GUID_WICPixelFormat ## fmt
-    static const struct test_data
-    {
-        const char *name;
-        const WICPixelFormatGUID *format;
-        BOOL src_valid;
-        BOOL dst_valid;
-        UINT dst_todo_count;
-        BOOL broken;
-    }
-    td[] =
-    {
-        {WIC_PIXEL_FORMAT(Undefined)},
-        {WIC_PIXEL_FORMAT(1bppIndexed), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(2bppIndexed), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(4bppIndexed), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(8bppIndexed), TRUE, TRUE, 26},
-        {WIC_PIXEL_FORMAT(BlackWhite), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(2bppGray), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(4bppGray), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(8bppGray), TRUE, TRUE, 26},
-        {WIC_PIXEL_FORMAT(16bppGray), TRUE, TRUE, 35},
-
-        {WIC_PIXEL_FORMAT(8bppAlpha), TRUE, TRUE, 35, TRUE},
-
-        {WIC_PIXEL_FORMAT(16bppBGR555), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(16bppBGR565), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(16bppBGRA5551), TRUE, TRUE, 33, TRUE},
-        {WIC_PIXEL_FORMAT(24bppBGR), TRUE, TRUE, 27},
-        {WIC_PIXEL_FORMAT(24bppRGB), TRUE, TRUE, 30},
-        {WIC_PIXEL_FORMAT(32bppBGR), TRUE, TRUE, 13},
-        {WIC_PIXEL_FORMAT(32bppBGRA), TRUE, TRUE, 13},
-        {WIC_PIXEL_FORMAT(32bppPBGRA), TRUE, TRUE, 13},
-        {WIC_PIXEL_FORMAT(32bppRGB), TRUE, TRUE, 11, TRUE},
-        {WIC_PIXEL_FORMAT(32bppRGBA), TRUE, TRUE, 11, TRUE},
-        {WIC_PIXEL_FORMAT(32bppPRGBA), TRUE, TRUE, 11, TRUE},
-        {WIC_PIXEL_FORMAT(32bppGrayFloat), TRUE, TRUE, 12},
-
-        {WIC_PIXEL_FORMAT(48bppRGB), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(48bppBGR), TRUE, TRUE, 35, TRUE},
-        {WIC_PIXEL_FORMAT(64bppRGB), TRUE, TRUE, 35, TRUE},
-        {WIC_PIXEL_FORMAT(64bppRGBA), TRUE, TRUE, 33},
-        {WIC_PIXEL_FORMAT(64bppBGRA), TRUE, TRUE, 35, TRUE},
-        {WIC_PIXEL_FORMAT(64bppPRGBA), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(64bppPBGRA), TRUE, TRUE, 35, TRUE},
-
-        {WIC_PIXEL_FORMAT(16bppGrayFixedPoint)},
-        {WIC_PIXEL_FORMAT(32bppBGR101010), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(48bppRGBFixedPoint)},
-        {WIC_PIXEL_FORMAT(48bppBGRFixedPoint)},
-        {WIC_PIXEL_FORMAT(96bppRGBFixedPoint)},
-        {WIC_PIXEL_FORMAT(96bppRGBFloat), TRUE, TRUE, 35, TRUE},
-        {WIC_PIXEL_FORMAT(128bppRGBAFloat), TRUE, TRUE, 33},
-        {WIC_PIXEL_FORMAT(128bppPRGBAFloat), TRUE, TRUE, 35},
-        {WIC_PIXEL_FORMAT(128bppRGBFloat), TRUE, TRUE, 33},
-
-        {WIC_PIXEL_FORMAT(32bppCMYK)},
-
-        {WIC_PIXEL_FORMAT(64bppRGBAFixedPoint)},
-        {WIC_PIXEL_FORMAT(64bppBGRAFixedPoint)},
-        {WIC_PIXEL_FORMAT(64bppRGBFixedPoint)},
-        {WIC_PIXEL_FORMAT(128bppRGBAFixedPoint)},
-        {WIC_PIXEL_FORMAT(128bppRGBFixedPoint)},
-
-        {WIC_PIXEL_FORMAT(64bppRGBAHalf)},
-        {WIC_PIXEL_FORMAT(64bppPRGBAHalf)},
-        {WIC_PIXEL_FORMAT(64bppRGBHalf)},
-        {WIC_PIXEL_FORMAT(48bppRGBHalf)},
-
-        {WIC_PIXEL_FORMAT(32bppRGBE)},
-
-        {WIC_PIXEL_FORMAT(16bppGrayHalf)},
-        {WIC_PIXEL_FORMAT(32bppGrayFixedPoint)},
-
-        {WIC_PIXEL_FORMAT(32bppRGBA1010102)},
-        {WIC_PIXEL_FORMAT(32bppRGBA1010102XR)},
-
-        /* Starting with Windows 10 v1809, this works as a source format */
-        {WIC_PIXEL_FORMAT(32bppR10G10B10A2), TRUE, FALSE, 0, TRUE},
-
-        {WIC_PIXEL_FORMAT(32bppR10G10B10A2HDR10)},
-
-        {WIC_PIXEL_FORMAT(64bppCMYK)},
-
-        {WIC_PIXEL_FORMAT(24bpp3Channels)},
-        {WIC_PIXEL_FORMAT(32bpp4Channels)},
-        {WIC_PIXEL_FORMAT(40bpp5Channels)},
-        {WIC_PIXEL_FORMAT(48bpp6Channels)},
-        {WIC_PIXEL_FORMAT(56bpp7Channels)},
-        {WIC_PIXEL_FORMAT(64bpp8Channels)},
-
-        {WIC_PIXEL_FORMAT(48bpp3Channels)},
-        {WIC_PIXEL_FORMAT(64bpp4Channels)},
-        {WIC_PIXEL_FORMAT(80bpp5Channels)},
-        {WIC_PIXEL_FORMAT(96bpp6Channels)},
-        {WIC_PIXEL_FORMAT(112bpp7Channels)},
-        {WIC_PIXEL_FORMAT(128bpp8Channels)},
-
-        {WIC_PIXEL_FORMAT(40bppCMYKAlpha)},
-        {WIC_PIXEL_FORMAT(80bppCMYKAlpha)},
-
-        {WIC_PIXEL_FORMAT(32bpp3ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(40bpp4ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(48bpp5ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(56bpp6ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(64bpp7ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(72bpp8ChannelsAlpha)},
-
-        {WIC_PIXEL_FORMAT(64bpp3ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(80bpp4ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(96bpp5ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(112bpp6ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(128bpp7ChannelsAlpha)},
-        {WIC_PIXEL_FORMAT(144bpp8ChannelsAlpha)},
-
-        {WIC_PIXEL_FORMAT(8bppY)},
-        {WIC_PIXEL_FORMAT(8bppCb)},
-        {WIC_PIXEL_FORMAT(8bppCr)},
-        {WIC_PIXEL_FORMAT(16bppCbCr)},
-
-        {WIC_PIXEL_FORMAT(16bppYQuantizedDctCoefficients)},
-        {WIC_PIXEL_FORMAT(16bppCbQuantizedDctCoefficients)},
-        {WIC_PIXEL_FORMAT(16bppCrQuantizedDctCoefficients)},
-    };
-#undef WIC_PIXEL_FORMAT
-    BOOL can_convert, can_native, can_broken;
-    HRESULT hr, expect_hr, broken_hr;
-    IWICFormatConverter *converter;
-    UINT todo_count;
-    UINT i, j;
-
-    hr = CoCreateInstance(&CLSID_WICDefaultFormatConverter, NULL, CLSCTX_INPROC_SERVER,
-        &IID_IWICFormatConverter, (void**)&converter);
-    ok(SUCCEEDED(hr), "CoCreateInstance failed, hr=%lx\n", hr);
-    if (FAILED(hr))
-        return;
-
-    for (j = 0; j < ARRAY_SIZE(td); j++)
-    {
-        todo_count = 0;
-
-        for (i = 0; i < ARRAY_SIZE(td); i++)
-        {
-            can_native = td[i].src_valid && td[j].dst_valid;
-            can_broken = !(td[i].broken || td[j].broken) && !can_native;
-            expect_hr = can_native ? S_OK : WINCODEC_ERR_UNSUPPORTEDPIXELFORMAT;
-            broken_hr = can_broken ? S_OK : WINCODEC_ERR_UNSUPPORTEDPIXELFORMAT;
-
-            can_convert = -1;
-            hr = IWICFormatConverter_CanConvert(converter, td[i].format, td[j].format, &can_convert);
-            todo_wine_if (can_native != can_convert) {
-                ok(hr == expect_hr || broken(hr == broken_hr),
-                        "CanConvert (%s -> %s) returned %lx\n", td[i].name, td[j].name, hr);
-                ok(can_convert == can_native || broken(can_convert == can_broken),
-                        "expected %i, got %i \n", can_native, can_convert);
-            }
-            /* Note that we ignore any conversions that Windows does not support */
-            if (can_native && !can_convert)
-                todo_count++;
-        }
-
-        todo_wine_if (td[j].dst_todo_count == todo_count && todo_count != 0)
-        ok(todo_count == 0 || broken(todo_count == 35 || todo_count == 11 || todo_count == 4 || todo_count == 1),
-            "CanConvert missing %d expected source formats to destination format %s.\n",
-            todo_count, td[j].name);
-    }
-
-    IWICFormatConverter_Release(converter);
-}
-
 static void test_converter_4bppGray(void)
 {
     BitmapTestSrc *src_obj;
@@ -1003,7 +723,7 @@
     {
         hr = IWICFormatConverter_CanConvert(converter, &GUID_WICPixelFormat32bppBGRA,
             &GUID_WICPixelFormat4bppGray, &can_convert);
-        todo_wine ok(SUCCEEDED(hr), "CanConvert returned %lx\n", hr);
+        ok(SUCCEEDED(hr), "CanConvert returned %lx\n", hr);
         todo_wine ok(can_convert, "expected TRUE, got %i\n", can_convert);
 
         hr = IWICFormatConverter_Initialize(converter, &src_obj->IWICBitmapSource_iface,
@@ -1063,30 +783,42 @@
     BOOL skippable;
 } property_opt_test_data;
 
+static const WCHAR wszTiffCompressionMethod[] = {'T','i','f','f','C','o','m','p','r','e','s','s','i','o','n','M','e','t','h','o','d',0};
+static const WCHAR wszCompressionQuality[] = {'C','o','m','p','r','e','s','s','i','o','n','Q','u','a','l','i','t','y',0};
+static const WCHAR wszInterlaceOption[] = {'I','n','t','e','r','l','a','c','e','O','p','t','i','o','n',0};
+static const WCHAR wszFilterOption[] = {'F','i','l','t','e','r','O','p','t','i','o','n',0};
+static const WCHAR wszImageQuality[] = {'I','m','a','g','e','Q','u','a','l','i','t','y',0};
+static const WCHAR wszBitmapTransform[] = {'B','i','t','m','a','p','T','r','a','n','s','f','o','r','m',0};
+static const WCHAR wszLuminance[] = {'L','u','m','i','n','a','n','c','e',0};
+static const WCHAR wszChrominance[] = {'C','h','r','o','m','i','n','a','n','c','e',0};
+static const WCHAR wszJpegYCrCbSubsampling[] = {'J','p','e','g','Y','C','r','C','b','S','u','b','s','a','m','p','l','i','n','g',0};
+static const WCHAR wszSuppressApp0[] = {'S','u','p','p','r','e','s','s','A','p','p','0',0};
+static const WCHAR wszEnableV5Header32bppBGRA[] = {'E','n','a','b','l','e','V','5','H','e','a','d','e','r','3','2','b','p','p','B','G','R','A',0};
+
 static const struct property_opt_test_data testdata_tiff_props[] = {
-    { L"TiffCompressionMethod", VT_UI1,         VT_UI1,  WICTiffCompressionDontCare },
-    { L"CompressionQuality",    VT_R4,          VT_EMPTY },
+    { wszTiffCompressionMethod, VT_UI1,         VT_UI1,  WICTiffCompressionDontCare },
+    { wszCompressionQuality,    VT_R4,          VT_EMPTY },
     { NULL }
 };
 
 static const struct property_opt_test_data testdata_png_props[] = {
-    { L"InterlaceOption", VT_BOOL, VT_BOOL, 0 },
-    { L"FilterOption",    VT_UI1,  VT_UI1, WICPngFilterUnspecified, 0.0f, TRUE /* not supported on XP/2k3 */},
+    { wszInterlaceOption, VT_BOOL, VT_BOOL, 0 },
+    { wszFilterOption,    VT_UI1,  VT_UI1, WICPngFilterUnspecified, 0.0f, TRUE /* not supported on XP/2k3 */},
     { NULL }
 };
 
 static const struct property_opt_test_data testdata_jpeg_props[] = {
-    { L"ImageQuality",         VT_R4,           VT_EMPTY },
-    { L"BitmapTransform",      VT_UI1,          VT_UI1, WICBitmapTransformRotate0 },
-    { L"Luminance",            VT_I4|VT_ARRAY,  VT_EMPTY },
-    { L"Chrominance",          VT_I4|VT_ARRAY,  VT_EMPTY },
-    { L"JpegYCrCbSubsampling", VT_UI1,          VT_UI1, WICJpegYCrCbSubsamplingDefault, 0.0f, TRUE }, /* not supported on XP/2k3 */
-    { L"SuppressApp0",         VT_BOOL,         VT_BOOL, FALSE },
+    { wszImageQuality,         VT_R4,           VT_EMPTY },
+    { wszBitmapTransform,      VT_UI1,          VT_UI1, WICBitmapTransformRotate0 },
+    { wszLuminance,            VT_I4|VT_ARRAY,  VT_EMPTY },
+    { wszChrominance,          VT_I4|VT_ARRAY,  VT_EMPTY },
+    { wszJpegYCrCbSubsampling, VT_UI1,          VT_UI1, WICJpegYCrCbSubsamplingDefault, 0.0f, TRUE }, /* not supported on XP/2k3 */
+    { wszSuppressApp0,         VT_BOOL,         VT_BOOL, FALSE },
     { NULL }
 };
 
 static const struct property_opt_test_data testdata_bmp_props[] = {
-    { L"EnableV5Header32bppBGRA", VT_BOOL, VT_BOOL, VARIANT_FALSE, 0.0f, TRUE }, /* Supported since Win7 */
+    { wszEnableV5Header32bppBGRA, VT_BOOL, VT_BOOL, VARIANT_FALSE, 0.0f, TRUE }, /* Supported since Win7 */
     { NULL }
 };
 
@@ -1271,7 +1003,7 @@
     ok(tiff.version == 42, "wrong TIFF version %u\n", tiff.version);
 
     pos.QuadPart = tiff.dir_offset;
-    hr = IStream_Seek(stream, pos, STREAM_SEEK_SET, NULL);
+    hr = IStream_Seek(stream, pos, SEEK_SET, NULL);
     ok(hr == S_OK, "IStream_Seek error %#lx\n", hr);
 
     hr = CoCreateInstance(&CLSID_WICIfdMetadataReader, NULL, CLSCTX_INPROC_SERVER,
@@ -1527,24 +1259,13 @@
         ok(png.filter == 0, "wrong filter %d\n", png.filter);
         ok(png.interlace == 0 || png.interlace == 1, "wrong interlace %d\n", png.interlace);
     }
-    else if (IsEqualGUID(format, &GUID_WICPixelFormat64bppRGBA))
-    {
-        ok(be_uint(png.width) == 32, "wrong width %u\n", be_uint(png.width));
-        ok(be_uint(png.height) == 2, "wrong height %u\n", be_uint(png.height));
-
-        ok(png.bit_depth == 16, "wrong bit_depth %d\n", png.bit_depth);
-        ok(png.color_type == 6, "wrong color_type %d\n", png.color_type);
-        ok(png.compression == 0, "wrong compression %d\n", png.compression);
-        ok(png.filter == 0, "wrong filter %d\n", png.filter);
-        ok(png.interlace == 0, "wrong interlace %d\n", png.interlace);
-    }
     else
         ok(0, "unknown PNG pixel format %s\n", wine_dbgstr_guid(format));
 }
 
 static void check_gif_format(IStream *stream, const WICPixelFormatGUID *format)
 {
-#pragma pack(push,1)
+#include "pshpack1.h"
     struct logical_screen_descriptor
     {
         char signature[6];
@@ -1559,7 +1280,7 @@
         BYTE background_color_index;
         BYTE pixel_aspect_ratio;
     } lsd;
-#pragma pack(pop)
+#include "poppack.h"
     UINT color_resolution;
     HRESULT hr;
 
@@ -1582,7 +1303,7 @@
     LARGE_INTEGER pos;
 
     pos.QuadPart = 0;
-    hr = IStream_Seek(stream, pos, STREAM_SEEK_SET, (ULARGE_INTEGER *)&pos);
+    hr = IStream_Seek(stream, pos, SEEK_SET, (ULARGE_INTEGER *)&pos);
     ok(hr == S_OK, "IStream_Seek error %#lx\n", hr);
 
     if (IsEqualGUID(encoder, &CLSID_WICPngEncoder))
@@ -1596,7 +1317,7 @@
     else
         ok(0, "unknown encoder %s\n", wine_dbgstr_guid(encoder));
 
-    hr = IStream_Seek(stream, pos, STREAM_SEEK_SET, NULL);
+    hr = IStream_Seek(stream, pos, SEEK_SET, NULL);
     ok(hr == S_OK, "IStream_Seek error %#lx\n", hr);
 }
 
@@ -2099,7 +1820,7 @@
     NULL};
 
 static const struct setting png_interlace_settings[] = {
-    {L"InterlaceOption", PROPBAG2_TYPE_DATA, VT_BOOL, (void*)VARIANT_TRUE},
+    {wszInterlaceOption, PROPBAG2_TYPE_DATA, VT_BOOL, (void*)VARIANT_TRUE},
     {NULL}
 };
 
@@ -2323,8 +2044,7 @@
     test_conversion(&testdata_24bppRGB, &testdata_4bppIndexed, "24bppRGB -> 4bppIndexed", TRUE);
     test_conversion(&testdata_24bppRGB, &testdata_8bppIndexed, "24bppRGB -> 8bppIndexed", FALSE);
 
-    test_conversion(&testdata_BlackWhite, &testdata_8bppIndexed_BW, "BlackWhite -> 8bppIndexed", FALSE);
-    test_conversion(&testdata_BlackWhite, &testdata_24bppBGR_BW, "BlackWhite -> 24bppBGR", FALSE);
+    test_conversion(&testdata_BlackWhite, &testdata_8bppIndexed_BW, "BlackWhite -> 8bppIndexed", TRUE);
     test_conversion(&testdata_1bppIndexed, &testdata_8bppIndexed_BW, "1bppIndexed -> 8bppIndexed", TRUE);
     test_conversion(&testdata_2bppIndexed, &testdata_8bppIndexed_4colors, "2bppIndexed -> 8bppIndexed", TRUE);
     test_conversion(&testdata_4bppIndexed, &testdata_8bppIndexed, "4bppIndexed -> 8bppIndexed", TRUE);
@@ -2366,22 +2086,8 @@
     test_conversion(&testdata_32bppBGRA, &testdata_16bppBGRA5551, "32bppBGRA -> 16bppBGRA5551", FALSE);
     test_conversion(&testdata_48bppRGB, &testdata_64bppRGBA_2, "48bppRGB -> 64bppRGBA", FALSE);
 
-    test_conversion(&testdata_48bppRGB, &testdata_128bppRGBFloat, "48bppRGB -> 128bppRGBFloat", FALSE);
-    test_conversion(&testdata_24bppBGR_2, &testdata_128bppRGBAFloat, "24bppBGR -> 128bppRGBAFloat", FALSE);
-    test_conversion(&testdata_32bppBGRA_2, &testdata_128bppRGBAFloat_2, "32bppBGRA -> 128bppRGBAFloat", FALSE);
-    test_conversion(&testdata_96bppRGBFloat, &testdata_128bppRGBFloat, "96bppRGBFloat -> 128bppRGBFloat", FALSE);
-    test_conversion(&testdata_96bppRGBFloat_2, &testdata_32bppBGRA_3, "96bppRGBFloat -> 32bppBGRA", FALSE);
-    test_conversion(&testdata_128bppRGBAFloat_2, &testdata_32bppBGRA_2, "128bppRGBAFloat -> 32bppBGRA", FALSE);
-
-    test_conversion(&testdata_48bppRGBHalf, &testdata_32bppBGRA_3, "48bppRGBHalf -> 32bppBGRA", FALSE);
-    test_conversion(&testdata_48bppRGBHalf, &testdata_128bppRGBFloat_2, "48bppRGBHalf -> 128bppRGBFloat", FALSE);
-
-    test_conversion(&testdata_16bppGrayHalf, &testdata_32bppBGRA_4, "16bppGrayHalf -> 32bppBGRA", FALSE);
-    test_conversion(&testdata_16bppGrayHalf, &testdata_128bppRGBFloat_3, "16bppGrayHalf -> 128bppRGBFloat", FALSE);
-
     test_invalid_conversion();
     test_default_converter();
-    test_can_convert();
     test_converter_4bppGray();
     test_converter_8bppGray();
     test_converter_8bppIndexed();
@@ -2406,8 +2112,6 @@
     test_encoder(&testdata_32bppBGR, &CLSID_WICPngEncoder,
                  &testdata_24bppBGR, &CLSID_WICPngDecoder, "PNG encoder 32bppBGR");
 }
-    test_encoder(&testdata_64bppRGBA, &CLSID_WICPngEncoder,
-                 &testdata_64bppRGBA, &CLSID_WICPngDecoder, "PNG encoder 64bppRGBA");
 
     test_encoder(&testdata_BlackWhite, &CLSID_WICBmpEncoder,
                  &testdata_1bppIndexed, &CLSID_WICBmpDecoder, "BMP encoder BlackWhite");
