diff --git a/dlls/ws2_32/tests/sock.c b/dlls/ws2_32/tests/sock.c
--- a/dlls/ws2_32/tests/sock.c
+++ b/dlls/ws2_32/tests/sock.c
@@ -26,19 +26,12 @@
 #include <winsock2.h>
 #include <windows.h>
 #include <winternl.h>
-#include <winioctl.h>
-#include <ddk/ntifs.h>
 #include <iphlpapi.h>
 #include <ws2tcpip.h>
 #include <wsipx.h>
-#include <afunix.h>
 #include <wsnwlink.h>
 #include <mswsock.h>
 #include <mstcpip.h>
-#include <bthsdpdef.h>
-#include <bluetoothapis.h>
-#include <bthdef.h>
-#include <ws2bth.h>
 #include <stdio.h>
 #include "wine/test.h"
 
@@ -3386,7 +3379,8 @@
     {
         SetLastError( 0xdeadbeef );
         sock = WSASocketA( tests[i].family, tests[i].type, tests[i].protocol, NULL, 0, 0 );
-        ok(WSAGetLastError() == tests[i].error, "Test %u: got wrong error %u\n", i, WSAGetLastError());
+        todo_wine_if (i == 7)
+            ok(WSAGetLastError() == tests[i].error, "Test %u: got wrong error %u\n", i, WSAGetLastError());
         if (tests[i].error)
         {
             ok(sock == INVALID_SOCKET, "Test %u: expected failure\n", i);
@@ -3717,51 +3711,6 @@
 
           closesocket(sock);
         }
-    }
-
-    /* Bluetooth RFCOMM socket tests */
-    SetLastError(0xdeadbeef);
-    sock = WSASocketA(AF_BTH, SOCK_STREAM, BTHPROTO_RFCOMM, NULL, 0, 0);
-    if (sock == INVALID_SOCKET)
-    {
-        err = WSAGetLastError();
-        ok(err == WSAEAFNOSUPPORT || err == WSAEPROTONOSUPPORT, "got error %d\n", err);
-        skip("Bluetooth is not supported\n");
-    }
-    else
-    {
-        WSAPROTOCOL_INFOA info;
-        closesocket(sock);
-
-        sock = WSASocketA(0, SOCK_STREAM, BTHPROTO_RFCOMM, NULL, 0, 0 );
-        ok(sock != INVALID_SOCKET, "Failed to create socket: %d\n", WSAGetLastError());
-
-        size = sizeof(socktype);
-        socktype = 0xdead;
-        err = getsockopt(sock, SOL_SOCKET, SO_TYPE, (char *) &socktype, &size);
-        ok(!err,"getsockopt failed with %d\n", WSAGetLastError());
-        ok(socktype == SOCK_STREAM, "Wrong socket type, expected %d received %d\n",
-           SOCK_STREAM, socktype);
-
-        size = sizeof(WSAPROTOCOL_INFOA);
-        err = getsockopt(sock, SOL_SOCKET, SO_PROTOCOL_INFOA, (char *) &info, &size);
-        ok(!err,"getsockopt failed with %d\n", WSAGetLastError());
-        ok(info.iProtocol == BTHPROTO_RFCOMM, "expected protocol %d, received %d\n",
-           BTHPROTO_RFCOMM, info.iProtocol);
-        ok(info.iAddressFamily == AF_BTH, "expected family %d, received %d\n",
-           AF_IPX, info.iProtocol);
-        ok(info.iSocketType == SOCK_STREAM, "expected type %d, received %d\n",
-           SOCK_DGRAM, info.iSocketType);
-        closesocket(sock);
-
-        /* SOCK_DGRAM is not supported by Bluetooth */
-        SetLastError(0xdeadbeef);
-        ok(WSASocketA(AF_BTH, SOCK_DGRAM, BTHPROTO_RFCOMM, NULL, 0, 0) == INVALID_SOCKET,
-           "WSASocketA should have failed\n");
-        err = WSAGetLastError();
-        ok(err == WSAEAFNOSUPPORT, "Expected 10047, received %d\n", err);
-
-        closesocket(sock);
     }
 }
 
@@ -4063,11 +4012,8 @@
     ret = WSAConnectByNameA(s, "winehq.org", "https", NULL, NULL, NULL, NULL, NULL, NULL);
     err = WSAGetLastError();
     ok(!ret, "WSAConnectByNameA should have failed\n");
-    ok(err == WSAEINVAL ||
-       err == WSAEFAULT ||   /* win10 >= 1809 */
-       err == WSAEOPNOTSUPP, /* win7, win8, win10 <= 1507 */
-       "expected error %u (WSAEINVAL) or %u (WSAEFAULT) or %u (WSAEOPNOTSUPP), got %u\n",
-       WSAEINVAL, WSAEFAULT, WSAEOPNOTSUPP, err);
+    ok(err == WSAEINVAL || err == WSAEFAULT, "expected error %u (WSAEINVAL) or %u (WSAEFAULT), got %u\n",
+       WSAEINVAL, WSAEFAULT, err); /* WSAEFAULT win10 >= 1809 */
     closesocket(s);
 
     /* Passing non-null as the reserved parameter */
@@ -6463,7 +6409,6 @@
             ok_(__FILE__, line)(msg.wParam == ctx->socket, "got wparam %#Ix\n", msg.wParam);
             todo_wine_if (todo_msg) ok_(__FILE__, line)(msg.lParam == flag2, "got second event %#Ix\n", msg.lParam);
         }
-        MsgWaitForMultipleObjects(0, NULL, FALSE, 10, QS_POSTMESSAGE);
         ret = PeekMessageA(&msg, ctx->window, WM_USER, WM_USER, PM_REMOVE);
         todo_wine_if (todo_msg && ret) ok_(__FILE__, line)(!ret, "got unexpected event %#Ix\n", msg.lParam);
         if (ret) any_fail = TRUE;
@@ -12819,45 +12764,6 @@
     }
 
     free(adapters);
-}
-
-static void test_bind_bluetooth(void)
-{
-    SOCKADDR_BTH addr = {0};
-    SOCKET sock, sock2;
-    INT err, ret;
-
-    sock = socket(AF_BTH, SOCK_STREAM, BTHPROTO_RFCOMM);
-    err = WSAGetLastError();
-    if (sock == INVALID_SOCKET)
-    {
-        ok(err == WSAEAFNOSUPPORT || err == WSAEPROTONOSUPPORT, "got error %d\n", err);
-        skip("Bluetooth is not supported\n");
-        return;
-    }
-
-    addr.addressFamily = AF_BTH;
-    addr.btAddr = 0;
-    addr.port = 200;
-    ret = bind(sock, (struct sockaddr *)&addr, sizeof(addr));
-    err = WSAGetLastError();
-    ok(ret == -1, "expected bind to fail\n");
-    ok(err == WSAEADDRNOTAVAIL || err == WSAENETDOWN, "got error %d\n", err);
-
-    addr.port = 20;
-    ret = bind(sock, (struct sockaddr *)&addr, sizeof(addr));
-    err = WSAGetLastError();
-    ok(!ret || err == WSAENETDOWN, "got error %d\n", err);
-
-    sock2 = socket(AF_BTH, SOCK_STREAM, BTHPROTO_RFCOMM);
-    ok(sock2 != INVALID_SOCKET, "got error %d\n", WSAGetLastError());
-    addr.port = BT_PORT_ANY;
-    ret = bind(sock2, (struct sockaddr *)&addr, sizeof(addr));
-    err = WSAGetLastError();
-    ok(!ret || err == WSAENETDOWN, "got error %d\n", err);
-
-    closesocket(sock);
-    closesocket(sock2);
 }
 
 /* Test calling methods on a socket which is currently connecting. */
@@ -13298,7 +13204,7 @@
     ret = send(server, "data", 4, 0);
     ok(ret == 4, "got %d\n", ret);
 
-    ret = WaitForSingleObject(thread, 2000);
+    ret = WaitForSingleObject(thread, 200);
     ok(!ret, "wait timed out\n");
     CloseHandle(thread);
 
@@ -13314,7 +13220,7 @@
     ret = send(server, "data", 4, 0);
     ok(ret == 4, "got %d\n", ret);
 
-    ret = WaitForSingleObject(thread, 2000);
+    ret = WaitForSingleObject(thread, 200);
     ok(!ret, "wait timed out\n");
     CloseHandle(thread);
 
@@ -14554,333 +14460,6 @@
     closesocket(client);
 }
 
-static void test_valid_handle(void)
-{
-    HANDLE duplicated, invalid;
-    SOCKET client, server;
-    char buffer[1];
-    WSABUF wsabuf;
-    DWORD size;
-    int ret;
-
-    /* Unlike WSAGetOverlappedResult(), duplicated handles are allowed. */
-
-    tcp_socketpair(&client, &server);
-    ret = DuplicateHandle(GetCurrentProcess(), (HANDLE)client,
-            GetCurrentProcess(), &duplicated, 0, FALSE, DUPLICATE_SAME_ACCESS);
-    ok(ret, "got error %lu\n", GetLastError());
-    invalid = CreateEventA(NULL, TRUE, TRUE, NULL);
-
-    ret = send((SOCKET)duplicated, buffer, 1, 0);
-    ok(ret == 1, "got %d\n", ret);
-
-    ret = sendto((SOCKET)duplicated, buffer, 1, 0, NULL, 0);
-    ok(ret == 1, "got %d\n", ret);
-
-    wsabuf.buf = buffer;
-    wsabuf.len = 1;
-
-    ret = WSASend((SOCKET)duplicated, &wsabuf, 1, NULL, 0, NULL, NULL);
-    ok(ret == -1, "got %d\n", ret);
-    ok(WSAGetLastError() == WSAEFAULT, "got error %u\n", WSAGetLastError());
-
-    ret = WSASend((SOCKET)duplicated, &wsabuf, 1, &size, 0, NULL, NULL);
-    ok(!ret, "got %d\n", ret);
-
-    ret = WSASend((SOCKET)invalid, &wsabuf, 1, NULL, 0, NULL, NULL);
-    ok(ret == -1, "got %d\n", ret);
-    ok(WSAGetLastError() == WSAENOTSOCK, "got error %u\n", WSAGetLastError());
-
-    ret = WSASend(INVALID_SOCKET, &wsabuf, 1, NULL, 0, NULL, NULL);
-    ok(ret == -1, "got %d\n", ret);
-    ok(WSAGetLastError() == WSAENOTSOCK, "got error %u\n", WSAGetLastError());
-
-    CloseHandle(invalid);
-    CloseHandle(duplicated);
-    closesocket(client);
-    closesocket(server);
-}
-
-static void test_afunix_path( const char *path )
-{
-    SOCKET listener, client, server = 0;
-    SOCKADDR_UN addr = { AF_UNIX }, empty = { AF_UNIX };
-    char buffer[sizeof(SOCKADDR_UN) * 2];
-    SOCKADDR_UN *out_addr = (SOCKADDR_UN *)buffer;
-    ULONG zero = 0, one = 1;
-    int size, ret;
-    DWORD attr;
-    HANDLE handle;
-
-    winetest_push_context( "%s", path );
-    strcpy(addr.sun_path, path);
-    DeleteFileA(addr.sun_path);  /* make sure it doesn't exist */
-
-    listener = socket(AF_UNIX, SOCK_STREAM, 0);
-    ok(listener != INVALID_SOCKET, "Could not create Unix socket: %lu\n", GetLastError());
-
-    ret = bind(listener, (SOCKADDR *)&addr, sizeof(addr));
-    ok(!ret, "Could not bind Unix socket: %lu\n", GetLastError());
-    attr = GetFileAttributesA(path);
-    ok( attr == (FILE_ATTRIBUTE_REPARSE_POINT | FILE_ATTRIBUTE_ARCHIVE), "wrong attr %lx\n", attr );
-
-    ret = listen(listener, 1);
-    ok(!ret, "Could not listen on Unix socket: %lu\n", GetLastError());
-
-    client = socket(AF_UNIX, SOCK_STREAM, 0);
-    ok(client != INVALID_SOCKET, "Failed to create second Unix socket: %lu\n", GetLastError());
-
-    ret = ioctlsocket(client, FIONBIO, &one);
-    ok(!ret, "Could not set AF_UNIX socket to nonblocking: %lu; skipping connection\n", GetLastError());
-    if (!ret)
-    {
-        ret = connect(client, (SOCKADDR *)&addr, sizeof(addr));
-        ok(!ret || (ret == SOCKET_ERROR && GetLastError() == WSAEWOULDBLOCK),
-            "Error when connecting to Unix socket: %lu\n", GetLastError());
-        server = accept(listener, NULL, NULL);
-        ok(server != INVALID_SOCKET, "Could not accept Unix socket connection: %lu\n", GetLastError());
-        ret = ioctlsocket(client, FIONBIO, &zero);
-        ok(!ret, "Could not set AF_UNIX socket to blocking: %lu\n", GetLastError());
-    }
-
-    memset(buffer, 0x55, sizeof(buffer));
-    size = sizeof(buffer);
-    ret = getsockname(listener, (SOCKADDR *)buffer, &size);
-    ok(!ret, "Could not get info on Unix socket: %lu\n", GetLastError());
-    ok(out_addr->sun_family == AF_UNIX, "wrong family %u\n", out_addr->sun_family);
-    ok(size == sizeof(addr.sun_family) + strlen(addr.sun_path) + 1, "wrong size %d\n", size );
-    ok(!strcmp(addr.sun_path, out_addr->sun_path), "wrong path %s\n", debugstr_a(out_addr->sun_path));
-    ok(buffer[size] == 0x55, "buffer overflow %x\n", buffer[size] );
-
-    memset(buffer, 0x55, sizeof(buffer));
-    size = sizeof(buffer);
-    ret = getsockname(client, (SOCKADDR *)buffer, &size);
-    ok(!ret, "Could not get info on Unix socket: %lu\n", GetLastError());
-    ok(out_addr->sun_family == AF_UNIX, "wrong family %u\n", out_addr->sun_family);
-    ok(size == sizeof(addr), "wrong size %d\n", size );
-    ok(!memcmp(empty.sun_path, out_addr->sun_path, sizeof(empty.sun_path)),
-        "wrong path %s\n", debugstr_a(out_addr->sun_path));
-
-    memset(buffer, 0x55, sizeof(buffer));
-    size = sizeof(buffer);
-    ret = getsockname(server, (SOCKADDR *)buffer, &size);
-    ok(!ret, "Could not get info on Unix socket: %lu\n", GetLastError());
-    ok(out_addr->sun_family == AF_UNIX, "wrong family %u\n", out_addr->sun_family);
-    ok(size == sizeof(addr.sun_family) + strlen(addr.sun_path) + 1, "wrong size %d\n", size );
-    ok(!strcmp(addr.sun_path, out_addr->sun_path), "wrong path %s\n", debugstr_a(out_addr->sun_path));
-    ok(buffer[size] == 0x55, "buffer overflow %x\n", buffer[size] );
-
-    memset(buffer, 0x55, sizeof(buffer));
-    size = sizeof(buffer);
-    ret = getpeername(listener, (SOCKADDR *)buffer, &size);
-    ok(ret == -1, "Got info on Unix socket: %lu\n", GetLastError());
-    ok(GetLastError() == WSAENOTCONN, "Incorrect error from getpeername: %ld\n", GetLastError());
-    ok(buffer[0] == 0x55, "getpeername returned incorrect path %s\n", debugstr_a(buffer));
-    ok(size == sizeof(buffer), "getpeername returned incorrect size %d\n", size);
-
-    memset(buffer, 0x55, sizeof(buffer));
-    size = sizeof(buffer);
-    ret = getpeername(client, (SOCKADDR *)buffer, &size);
-    ok(!ret, "Could not get info on Unix socket: %lu\n", GetLastError());
-    ok(out_addr->sun_family == AF_UNIX, "wrong family %u\n", out_addr->sun_family);
-    ok(size == sizeof(addr), "wrong size %d\n", size );
-    ok(!memcmp(addr.sun_path, out_addr->sun_path, sizeof(addr.sun_path)),
-        "wrong path %s\n", debugstr_a(out_addr->sun_path));
-
-    memset(buffer, 0x55, sizeof(buffer));
-    size = sizeof(buffer);
-    ret = getpeername(server, (SOCKADDR *)buffer, &size);
-    ok(!ret, "Could not get info on Unix socket: %lu\n", GetLastError());
-    ok(out_addr->sun_family == AF_UNIX, "wrong family %u\n", out_addr->sun_family);
-    ok(size == sizeof(addr), "wrong size %d\n", size );
-    ok(!memcmp(empty.sun_path, out_addr->sun_path, sizeof(empty.sun_path)),
-        "wrong path %s\n", debugstr_a(out_addr->sun_path));
-
-    closesocket(listener);
-    closesocket(client);
-    closesocket(server);
-
-    attr = GetFileAttributesA(path);
-    ok( attr == (FILE_ATTRIBUTE_REPARSE_POINT | FILE_ATTRIBUTE_ARCHIVE) ||
-        broken(attr == FILE_ATTRIBUTE_ARCHIVE), /* win10 1809 */ "wrong attr %lx\n", attr );
-
-    if (attr & FILE_ATTRIBUTE_REPARSE_POINT)
-    {
-        REPARSE_DATA_BUFFER *data = (REPARSE_DATA_BUFFER *)buffer;
-        handle = CreateFileA( path, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
-                              NULL, OPEN_EXISTING, 0, NULL );
-        ok( handle == INVALID_HANDLE_VALUE, "open succeeded\n" );
-        ok( GetLastError() == ERROR_CANT_ACCESS_FILE, "wrong error %lu\n", GetLastError() );
-
-        handle = CreateFileA( path, GENERIC_READ, FILE_SHARE_READ | FILE_SHARE_WRITE,
-                              NULL, OPEN_EXISTING, FILE_FLAG_OPEN_REPARSE_POINT, NULL );
-        ok( handle != INVALID_HANDLE_VALUE, "open failed %lu\n", GetLastError() );
-        ret = DeviceIoControl( handle, FSCTL_GET_REPARSE_POINT, NULL, 0,
-                               buffer, sizeof(buffer), NULL, NULL );
-        ok( ret, "DeviceIoControl failed %lu\n", GetLastError() );
-        ok( data->ReparseTag == IO_REPARSE_TAG_AF_UNIX, "got tag %lx\n", data->ReparseTag );
-        ok( !data->ReparseDataLength, "got len %u\n", data->ReparseDataLength );
-        CloseHandle( handle );
-    }
-
-    ret = DeleteFileA(path);
-    ok(ret, "DeleteFileA on socket file failed: %lu\n", GetLastError());
-    attr = GetFileAttributesA(path);
-    ok( attr == INVALID_FILE_ATTRIBUTES, "wrong attr %lx\n", attr );
-
-    winetest_pop_context();
-}
-
-static void test_afunix(void)
-{
-    SOCKET listener, client, server = 0;
-    SOCKADDR_UN addr = { AF_UNIX, "test_afunix.sock" };
-    char serverBuf[] = "ws2_32/AF_UNIX socket test";
-    char clientBuf[sizeof(serverBuf)] = { 0 };
-    char path[MAX_PATH];
-    WCHAR pathW[MAX_PATH];
-    UNICODE_STRING ntPath;
-    SOCKADDR_UN outAddr = { 0 };
-    int outAddrSize = sizeof(outAddr);
-    SOCKADDR_UN truncatedAddr = { 0 };
-    ULONG zero = 0;
-    ULONG one = 1;
-    int ret;
-
-    /* Test connection and send/recv */
-    listener = socket(AF_UNIX, SOCK_STREAM, 0);
-    if (listener == INVALID_SOCKET && GetLastError() == WSAEAFNOSUPPORT)
-    {
-        todo_wine
-        win_skip("AF_UNIX sockets not supported\n");
-        return;
-    }
-
-    ok(listener != INVALID_SOCKET, "Could not create Unix socket: %lu\n", GetLastError());
-    ret = bind(listener, (SOCKADDR *)&addr, 0);
-    ok(ret && GetLastError() == WSAEFAULT, "Incorrect error: %lu\n", GetLastError());
-    ret = bind(listener, (SOCKADDR *)&addr, 2);
-    ok(!ret, "Could not bind Unix socket: %lu\n", GetLastError());
-    ret = listen(listener, 1);
-    ok(!ret, "Could not listen on Unix socket: %lu\n", GetLastError());
-    closesocket(listener);
-
-    listener = socket(AF_UNIX, SOCK_STREAM, 0);
-    ok(listener != INVALID_SOCKET, "Could not create Unix socket: %lu\n", GetLastError());
-    ret = bind(listener, (SOCKADDR *)&addr, 3);
-    ok(!ret, "Could not bind Unix socket: %lu\n", GetLastError());
-
-    memcpy(&truncatedAddr, &addr, 3);
-    ret = getsockname(listener, (SOCKADDR *)&outAddr, &outAddrSize);
-    ok(!ret, "Could not get info on Unix socket: %lu\n", GetLastError());
-    ok(!memcmp(truncatedAddr.sun_path, outAddr.sun_path, sizeof(addr.sun_path)),
-        "getsockname returned incorrect path %s  / %s\n",
-        debugstr_a(outAddr.sun_path), debugstr_a(truncatedAddr.sun_path));
-    ok(outAddrSize == sizeof(outAddr.sun_family) + strlen(outAddr.sun_path) + 1,
-        "getsockname returned incorrect size %d for %s\n", outAddrSize, debugstr_a(truncatedAddr.sun_path));
-    closesocket(listener);
-    ret = DeleteFileA(truncatedAddr.sun_path);
-    ok(ret, "DeleteFileA on socket file failed: %lu\n", GetLastError());
-    ok(GetFileAttributesA(truncatedAddr.sun_path) == INVALID_FILE_ATTRIBUTES,
-        "%s still exists\n", debugstr_a(truncatedAddr.sun_path));
-
-    listener = socket(AF_UNIX, SOCK_STREAM, 0);
-    ok(listener != INVALID_SOCKET, "Could not create Unix socket: %lu\n", GetLastError());
-
-    ret = bind(listener, (SOCKADDR *)&addr, sizeof(SOCKADDR_UN));
-    ok(!ret, "Could not bind Unix socket: %lu\n", GetLastError());
-
-    ret = listen(listener, 1);
-    ok(!ret, "Could not listen on Unix socket: %lu\n", GetLastError());
-
-    client = socket(AF_UNIX, SOCK_STREAM, 0);
-    ok(client != INVALID_SOCKET, "Failed to create second Unix socket: %lu\n", GetLastError());
-
-    ret = ioctlsocket(client, FIONBIO, &one);
-    ok(!ret, "Could not set AF_UNIX socket to nonblocking: %lu; skipping connection\n", GetLastError());
-    if (!ret)
-    {
-        ret = connect(client, (SOCKADDR *)&addr, sizeof(addr));
-        ok(!ret || (ret == SOCKET_ERROR && GetLastError() == WSAEWOULDBLOCK),
-            "Error when connecting to Unix socket: %lu\n", GetLastError());
-        server = accept(listener, NULL, NULL);
-        ok(server != INVALID_SOCKET, "Could not accept Unix socket connection: %lu\n", GetLastError());
-        ret = ioctlsocket(client, FIONBIO, &zero);
-        ok(!ret, "Could not set AF_UNIX socket to blocking: %lu\n", GetLastError());
-    }
-
-    ret = send(server, serverBuf, sizeof(serverBuf), 0);
-    ok(ret == sizeof(serverBuf), "Incorrect return value from send: %d\n", ret);
-    ret = recv(client, clientBuf, sizeof(serverBuf), 0);
-    ok(ret == sizeof(serverBuf), "Incorrect return value from recv: %d\n", ret);
-    ok(!memcmp(serverBuf, clientBuf, sizeof(serverBuf)), "Data mismatch over Unix socket\n");
-
-    memset(clientBuf, 0, sizeof(clientBuf));
-
-    ret = sendto(server, serverBuf, sizeof(serverBuf), 0, NULL, 0);
-    ok(ret == sizeof(serverBuf), "Incorrect return value from sendto: %d\n", ret);
-    ret = recvfrom(client, clientBuf, sizeof(serverBuf), 0, NULL, 0);
-    ok(ret == sizeof(serverBuf), "Incorrect return value from recvfrom: %d\n", ret);
-    ok(!memcmp(serverBuf, clientBuf, sizeof(serverBuf)), "Data mismatch over Unix socket\n");
-
-    memset(serverBuf, 0, sizeof(serverBuf));
-
-    ret = send(client, clientBuf, sizeof(clientBuf), 0);
-    ok(ret == sizeof(clientBuf), "Incorrect return value from send: %d\n", ret);
-    ret = recv(server, serverBuf, sizeof(clientBuf), 0);
-    ok(ret == sizeof(serverBuf), "Incorrect return value from recv: %d\n", ret);
-    ok(!memcmp(serverBuf, clientBuf, sizeof(clientBuf)), "Data mismatch over Unix socket\n");
-
-    memset(serverBuf, 0, sizeof(serverBuf));
-
-    ret = sendto(client, clientBuf, sizeof(clientBuf), 0, NULL, 0);
-    ok(ret == sizeof(clientBuf), "Incorrect return value from sendto: %d\n", ret);
-    ret = recvfrom(server, serverBuf, sizeof(clientBuf), 0, NULL, 0);
-    ok(ret == sizeof(serverBuf), "Incorrect return value from recvfrom: %d\n", ret);
-    ok(!memcmp(serverBuf, clientBuf, sizeof(clientBuf)), "Data mismatch over Unix socket\n");
-
-    closesocket(listener);
-    closesocket(client);
-    closesocket(server);
-
-    /* Test socket file deletion */
-    ret = DeleteFileA(addr.sun_path);
-    ok(ret, "DeleteFileA on socket file failed: %lu\n", GetLastError());
-    ok(GetFileAttributesA(addr.sun_path) == INVALID_FILE_ATTRIBUTES &&
-        GetLastError() == ERROR_FILE_NOT_FOUND,
-        "Failed to delete socket file at path '%s'\n",
-        addr.sun_path);
-
-    /* Test failure modes */
-    listener = socket(AF_UNIX, SOCK_STREAM, 0);
-    ok(listener != INVALID_SOCKET, "Could not create Unix socket: %lu\n", GetLastError());
-    ret = bind(listener, (SOCKADDR *)&addr, sizeof(SOCKADDR_UN));
-    ok(!ret, "Could not bind Unix socket to path '%s': %lu\n", addr.sun_path, GetLastError());
-    closesocket(listener);
-    listener = socket(AF_UNIX, SOCK_STREAM, 0);
-    ok(listener != INVALID_SOCKET, "Could not create Unix socket: %lu\n", GetLastError());
-    ret = bind(listener, (SOCKADDR *)&addr, sizeof(SOCKADDR_UN));
-    ok(ret, "Bound path %s despite existing socket file\n", debugstr_a(addr.sun_path));
-    ok(GetLastError() == WSAEADDRINUSE, "wrong error %lu\n", GetLastError());
-    closesocket(listener);
-    ret = DeleteFileA(addr.sun_path);
-    ok(ret, "DeleteFileA on socket file failed: %lu\n", GetLastError());
-    ok(GetFileAttributesA(addr.sun_path) == INVALID_FILE_ATTRIBUTES,
-        "%s still exists\n", debugstr_a(addr.sun_path));
-
-    /* Test different path types (relative, NT, etc.) */
-    test_afunix_path( addr.sun_path );
-    test_afunix_path( ".\\tmp.sock" );
-    GetTempPathA(MAX_PATH, path);
-    strcat(path, "tmp.sock");
-    test_afunix_path( path );
-    GetTempPathW(MAX_PATH, pathW);
-    wcscat(pathW, L"tmp.sock");
-    RtlDosPathNameToNtPathName_U(pathW, &ntPath, NULL, NULL);
-    RtlUnicodeToMultiByteN(path, sizeof(addr.sun_path), NULL, ntPath.Buffer, ntPath.Length + sizeof(WCHAR));
-    test_afunix_path( path );
-}
-
 START_TEST( sock )
 {
     int i;
@@ -14953,7 +14532,6 @@
     test_connect_completion_port();
     test_shutdown_completion_port();
     test_bind();
-    test_bind_bluetooth();
     test_connecting_socket();
     test_WSAGetOverlappedResult();
     test_nonblocking_async_recv();
@@ -14967,8 +14545,6 @@
     test_tcp_sendto_recvfrom();
     test_broadcast();
     test_send_buffering();
-    test_valid_handle();
-    test_afunix();
 
     /* There is apparently an obscure interaction between this test and
      * test_WSAGetOverlappedResult().
