diff --git a/dlls/winegstreamer/quartz_parser.c b/dlls/winegstreamer/quartz_parser.c
--- a/dlls/winegstreamer/quartz_parser.c
+++ b/dlls/winegstreamer/quartz_parser.c
@@ -37,11 +37,11 @@
 WINE_DEFAULT_DEBUG_CHANNEL(quartz);
 
 static const GUID MEDIASUBTYPE_CVID = {mmioFOURCC('c','v','i','d'), 0x0000, 0x0010, {0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}};
-static const GUID MEDIASUBTYPE_IV50 = {mmioFOURCC('I','V','5','0'), 0x0000, 0x0010, {0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}};
 static const GUID MEDIASUBTYPE_VC1S = {mmioFOURCC('V','C','1','S'), 0x0000, 0x0010, {0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}};
 static const GUID MEDIASUBTYPE_MP3  = {WAVE_FORMAT_MPEGLAYER3, 0x0000, 0x0010, {0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}};
 static const GUID MEDIASUBTYPE_WMV_Unknown = {0x7ce12ca9, 0xbfbf, 0x43d9, {0x9d, 0x00, 0x82, 0xb8, 0xed, 0x54, 0x31, 0x6b}};
 DEFINE_GUID(MEDIASUBTYPE_ABGR32,D3DFMT_A8B8G8R8,0x524f,0x11ce,0x9f,0x53,0x00,0x20,0xaf,0x0b,0xa7,0x70);
+static const GUID MEDIASUBTYPE_XMAUDIO2 = {0x0166, 0x0000, 0x0010, {0x80, 0x00, 0x00, 0xaa, 0x00, 0x38, 0x9b, 0x71}};
 
 struct parser
 {
@@ -53,8 +53,7 @@
 
     struct parser_source **sources;
     unsigned int source_count;
-    bool enum_sink_first;
-    bool output_compressed;
+    BOOL enum_sink_first;
 
     wg_parser_t wg_parser;
 
@@ -88,8 +87,6 @@
     IQualityControl IQualityControl_iface;
 
     wg_parser_stream_t wg_stream;
-
-    struct wg_format current_format;
 
     SourceSeeking seek;
 
@@ -308,6 +305,11 @@
             subtype = &MEDIASUBTYPE_WMAUDIO2;
             codec_data_len = WMAUDIO2_WFX_EXTRA_BYTES;
             fmt_tag = WAVE_FORMAT_WMAUDIO2;
+            if (format->u.audio.is_xma)
+            {
+                subtype = &MEDIASUBTYPE_XMAUDIO2;
+                fmt_tag = 0x0166;
+            }
             break;
         case 3:
             subtype = &MEDIASUBTYPE_WMAUDIO3;
@@ -401,10 +403,6 @@
             /* Both ffmpeg's encoder and a Cinepak file seen in the wild report
              * 24 bpp. ffmpeg sets biSizeImage as below; others may be smaller,
              * but as long as every sample fits into our allocator, we're fine. */
-        case WG_MAJOR_TYPE_VIDEO_INDEO:
-            /* Similar situation, though we don't even have ffmpeg to comapre.
-             * Files from Close Combat 3 report 24 bpp, but a smaller
-             * biSizeImage. */
             return format->u.video.width * format->u.video.height * 3;
 
         case WG_MAJOR_TYPE_VIDEO_MPEG1:
@@ -469,6 +467,7 @@
 
         case WG_MAJOR_TYPE_AUDIO_MPEG4:
         case WG_MAJOR_TYPE_VIDEO_H264:
+        case WG_MAJOR_TYPE_VIDEO_INDEO:
             FIXME("Format %u not implemented!\n", format->major_type);
             return 0;
 
@@ -565,10 +564,7 @@
     uint32_t frame_time;
 
     if (format->u.video.format == WG_VIDEO_FORMAT_UNKNOWN)
-    {
-        WARN("Unknown video format.\n");
         return false;
-    }
 
     if (!(video_format = CoTaskMemAlloc(sizeof(*video_format))))
         return false;
@@ -625,43 +621,6 @@
     mt->majortype = MEDIATYPE_Video;
     mt->subtype = MEDIASUBTYPE_CVID;
     mt->bTemporalCompression = TRUE;
-    mt->lSampleSize = 1;
-    mt->formattype = FORMAT_VideoInfo;
-    mt->cbFormat = sizeof(*video_format);
-    mt->pbFormat = (BYTE *)video_format;
-
-    memset(video_format, 0, sizeof(*video_format));
-    if ((frame_time = MulDiv(10000000, format->u.video.fps_d, format->u.video.fps_n)) != -1)
-        video_format->AvgTimePerFrame = frame_time;
-    video_format->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
-    video_format->bmiHeader.biWidth = format->u.video.width;
-    video_format->bmiHeader.biHeight = format->u.video.height;
-    video_format->bmiHeader.biPlanes = 1;
-    video_format->bmiHeader.biBitCount = 24;
-    video_format->bmiHeader.biCompression = mt->subtype.Data1;
-    video_format->bmiHeader.biSizeImage = wg_format_get_max_size(format);
-
-    return true;
-}
-
-static bool amt_from_wg_format_video_indeo(AM_MEDIA_TYPE *mt, const struct wg_format *format)
-{
-    VIDEOINFOHEADER *video_format;
-    uint32_t frame_time;
-
-    if (!(video_format = CoTaskMemAlloc(sizeof(*video_format))))
-        return false;
-
-    mt->majortype = MEDIATYPE_Video;
-    if (format->u.video.version == 5)
-    {
-        mt->subtype = MEDIASUBTYPE_IV50;
-    }
-    else
-    {
-        FIXME("Unhandled version %u.\n", format->u.video.version);
-        return false;
-    }
     mt->lSampleSize = 1;
     mt->formattype = FORMAT_VideoInfo;
     mt->cbFormat = sizeof(*video_format);
@@ -779,6 +738,7 @@
     {
     case WG_MAJOR_TYPE_AUDIO_MPEG4:
     case WG_MAJOR_TYPE_VIDEO_H264:
+    case WG_MAJOR_TYPE_VIDEO_INDEO:
         FIXME("Format %u not implemented!\n", format->major_type);
         /* fallthrough */
     case WG_MAJOR_TYPE_UNKNOWN:
@@ -798,9 +758,6 @@
 
     case WG_MAJOR_TYPE_VIDEO_CINEPAK:
         return amt_from_wg_format_video_cinepak(mt, format);
-
-    case WG_MAJOR_TYPE_VIDEO_INDEO:
-        return amt_from_wg_format_video_indeo(mt, format);
 
     case WG_MAJOR_TYPE_VIDEO_WMV:
         return amt_from_wg_format_video_wmv(mt, format, wm);
@@ -926,7 +883,8 @@
     return true;
 }
 
-static bool amt_to_wg_format_audio_wma(const AM_MEDIA_TYPE *mt, struct wg_format *format)
+static bool amt_to_wg_format_audio_wma(const AM_MEDIA_TYPE *mt, struct wg_format *format,
+        bool is_xma)
 {
     const WAVEFORMATEX *audio_format = (const WAVEFORMATEX *)mt->pbFormat;
 
@@ -952,6 +910,7 @@
     else
         assert(false);
     format->major_type = WG_MAJOR_TYPE_AUDIO_WMA;
+    format->u.audio.is_xma = is_xma;
     format->u.audio.bitrate = audio_format->nAvgBytesPerSec * 8;
     format->u.audio.rate = audio_format->nSamplesPerSec;
     format->u.audio.depth = audio_format->wBitsPerSample;
@@ -1057,32 +1016,6 @@
     return true;
 }
 
-static bool amt_to_wg_format_video_indeo(const AM_MEDIA_TYPE *mt, struct wg_format *format)
-{
-    const VIDEOINFOHEADER *video_format = (const VIDEOINFOHEADER *)mt->pbFormat;
-
-    if (!IsEqualGUID(&mt->formattype, &FORMAT_VideoInfo))
-    {
-        FIXME("Unknown format type %s.\n", debugstr_guid(&mt->formattype));
-        return false;
-    }
-    if (mt->cbFormat < sizeof(VIDEOINFOHEADER) || !mt->pbFormat)
-    {
-        ERR("Unexpected format size %lu.\n", mt->cbFormat);
-        return false;
-    }
-
-    format->major_type = WG_MAJOR_TYPE_VIDEO_CINEPAK;
-    if (IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_IV50))
-        format->u.video.version = 5;
-    format->u.video.width = video_format->bmiHeader.biWidth;
-    format->u.video.height = video_format->bmiHeader.biHeight;
-    format->u.video.fps_n = 10000000;
-    format->u.video.fps_d = video_format->AvgTimePerFrame;
-
-    return true;
-}
-
 static bool amt_to_wg_format_video_wmv(const AM_MEDIA_TYPE *mt, struct wg_format *format)
 {
     const VIDEOINFOHEADER *video_format = (const VIDEOINFOHEADER *)mt->pbFormat;
@@ -1159,8 +1092,6 @@
     {
         if (IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_CVID))
             return amt_to_wg_format_video_cinepak(mt, format);
-        if (IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_IV50))
-            return amt_to_wg_format_video_indeo(mt, format);
         if (IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_WMV1)
                 || IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_WMV2)
                 || IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_WMVA)
@@ -1185,7 +1116,9 @@
                 || IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_WMAUDIO2)
                 || IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_WMAUDIO3)
                 || IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_WMAUDIO_LOSSLESS))
-            return amt_to_wg_format_audio_wma(mt, format);
+            return amt_to_wg_format_audio_wma(mt, format, false);
+        if (IsEqualGUID(&mt->subtype, &MEDIASUBTYPE_XMAUDIO2))
+            return amt_to_wg_format_audio_wma(mt, format, true);
         return amt_to_wg_format_audio(mt, format);
     }
 
@@ -1314,11 +1247,10 @@
 
 /* Send a single GStreamer buffer (splitting it into multiple IMediaSamples if
  * necessary). */
-static void send_buffer(struct parser_source *pin, struct parser *filter, const struct wg_parser_buffer *buffer)
+static void send_buffer(struct parser_source *pin, const struct wg_parser_buffer *buffer)
 {
     HRESULT hr;
     IMediaSample *sample;
-    AM_MEDIA_TYPE *mt;
 
     if (pin->need_segment)
     {
@@ -1345,16 +1277,6 @@
                 break;
             }
 
-            if ((hr = IMediaSample_GetMediaType(sample, &mt)) == S_OK)
-            {
-                FIXME("Dynamic format change.\n");
-                DeleteMediaType(mt);
-            }
-            else if (hr != S_FALSE)
-            {
-                ERR("Failed to get media type, hr %#lx.\n", hr);
-            }
-
             advance = min(IMediaSample_GetSize(sample), buffer->size - offset);
 
             hr = send_sample(pin, sample, buffer, offset, advance, format->nAvgBytesPerSec);
@@ -1375,64 +1297,7 @@
         }
         else
         {
-            /* Dynamic format change. */
-            if ((hr = IMediaSample_GetMediaType(sample, &mt)) == S_OK)
-            {
-                struct wg_format format;
-
-                if (filter->output_compressed)
-                {
-                    ERR("Ignoring dynamic format change attempt for compressed output.\n");
-                    send_sample(pin, sample, buffer, 0, buffer->size, 0);
-                }
-                else if (amt_to_wg_format(mt, &format))
-                {
-                    if (!memcmp(&format, &pin->current_format, sizeof(format)))
-                    {
-                        send_sample(pin, sample, buffer, 0, buffer->size, 0);
-                    }
-                    else
-                    {
-                        TRACE("Executing dynamic format change. Current format:\n");
-                        strmbase_dump_media_type(&pin->pin.pin.mt);
-                        TRACE("New format:\n");
-                        strmbase_dump_media_type(mt);
-
-                        FreeMediaType(&pin->pin.pin.mt);
-                        CopyMediaType(&pin->pin.pin.mt, mt);
-                        pin->current_format = format;
-                        wg_parser_stream_enable(pin->wg_stream, &format);
-
-                        /* We can't send the wg_parser_buffer we were about to
-                         * send; it's in the old format.
-                         *
-                         * Also, we need to seek to re-decode any further
-                         * queued buffers. This is not reliably seamless.
-                         * See reader_SetOutputProps() in wm_reader.c.
-                         * In practice it's unlikely to matter, since dynamic
-                         * reconnection is usually done only at the beginning
-                         * of a stream. */
-                        wg_parser_stream_seek(pin->wg_stream, pin->seek.dRate, pin->seek.llCurrent, 0,
-                                AM_SEEKING_AbsolutePositioning, AM_SEEKING_NoPositioning);
-                    }
-                }
-                else
-                {
-                    /* This isn't supposed to happen; the downstream filter
-                     * should call QueryAccept() first. */
-                    ERR("Attempt to dynamically set an unsupported format.\n");
-                }
-
-                DeleteMediaType(mt);
-            }
-            else if (hr == S_FALSE)
-            {
-                send_sample(pin, sample, buffer, 0, buffer->size, 0);
-            }
-            else
-            {
-                ERR("Failed to get media type, hr %#lx.\n", hr);
-            }
+            hr = send_sample(pin, sample, buffer, 0, buffer->size, 0);
 
             IMediaSample_Release(sample);
         }
@@ -1476,7 +1341,7 @@
 
         if (wg_parser_stream_get_buffer(filter->wg_parser, pin->wg_stream, &buffer))
         {
-            send_buffer(pin, filter, &buffer);
+            send_buffer(pin, &buffer);
         }
         else
         {
@@ -1613,13 +1478,14 @@
     for (i = 0; i < filter->source_count; ++i)
     {
         struct parser_source *source = filter->sources[i];
+        struct wg_format format;
         bool ret;
 
         if (source->pin.pin.peer)
         {
-            ret = amt_to_wg_format(&source->pin.pin.mt, &source->current_format);
+            ret = amt_to_wg_format(&source->pin.pin.mt, &format);
             assert(ret);
-            wg_parser_stream_enable(source->wg_stream, &source->current_format);
+            wg_parser_stream_enable(source->wg_stream, &format);
         }
         else
         {
@@ -1789,11 +1655,8 @@
     return amt_to_wg_format(mt, &format) ? S_OK : S_FALSE;
 }
 
-static HRESULT decodebin_parser_source_get_media_type(struct parser_source *pin,
-        unsigned int index, AM_MEDIA_TYPE *mt)
-{
-    struct wg_format format;
-
+static HRESULT get_raw_media_type(struct wg_format *format, unsigned int index, AM_MEDIA_TYPE *mt)
+{
     static const enum wg_video_format video_formats[] =
     {
         /* Try to prefer YUV formats over RGB ones. Most decoders output in the
@@ -1813,31 +1676,29 @@
         WG_VIDEO_FORMAT_RGB15,
     };
 
-    wg_parser_stream_get_current_format(pin->wg_stream, &format);
-
     memset(mt, 0, sizeof(AM_MEDIA_TYPE));
 
-    if (amt_from_wg_format(mt, &format, false))
+    if (amt_from_wg_format(mt, format, false))
     {
         if (!index--)
             return S_OK;
         FreeMediaType(mt);
     }
 
-    if (format.major_type == WG_MAJOR_TYPE_VIDEO && index < ARRAY_SIZE(video_formats))
-    {
-        format.u.video.format = video_formats[index];
+    if (format->major_type == WG_MAJOR_TYPE_VIDEO && index < ARRAY_SIZE(video_formats))
+    {
+        format->u.video.format = video_formats[index];
         /* Downstream filters probably expect RGB video to be bottom-up. */
-        if (format.u.video.height > 0 && wg_video_format_is_rgb(video_formats[index]))
-            format.u.video.height = -format.u.video.height;
-        if (!amt_from_wg_format(mt, &format, false))
+        if (format->u.video.height > 0 && wg_video_format_is_rgb(video_formats[index]))
+            format->u.video.height = -format->u.video.height;
+        if (!amt_from_wg_format(mt, format, false))
             return E_OUTOFMEMORY;
         return S_OK;
     }
-    else if (format.major_type == WG_MAJOR_TYPE_AUDIO && !index)
-    {
-        format.u.audio.format = WG_AUDIO_FORMAT_S16LE;
-        if (!amt_from_wg_format(mt, &format, false))
+    else if (format->major_type == WG_MAJOR_TYPE_AUDIO && !index)
+    {
+        format->u.audio.format = WG_AUDIO_FORMAT_S16LE;
+        if (!amt_from_wg_format(mt, format, false))
             return E_OUTOFMEMORY;
         return S_OK;
     }
@@ -1845,6 +1706,17 @@
     return VFW_S_NO_MORE_ITEMS;
 }
 
+static HRESULT decodebin_parser_source_get_media_type(struct parser_source *pin,
+        unsigned int index, AM_MEDIA_TYPE *mt)
+{
+    struct wg_format format;
+
+    TRACE("pin %p, index %d, mt %p.\n", pin, (int)index, mt);
+
+    wg_parser_stream_get_current_format(pin->wg_stream, &format);
+    return get_raw_media_type(&format, index, mt);
+}
+
 static HRESULT parser_create(BOOL output_compressed, struct parser **parser)
 {
     struct parser *object;
@@ -1852,9 +1724,7 @@
     if (!(object = calloc(1, sizeof(*object))))
         return E_OUTOFMEMORY;
 
-    object->output_compressed = output_compressed;
-
-    if (!(object->wg_parser = wg_parser_create(output_compressed)))
+    if (!(object->wg_parser = wg_parser_create(output_compressed, FALSE)))
     {
         free(object);
         return E_OUTOFMEMORY;
@@ -2423,6 +2293,8 @@
     HRESULT hr;
 
     wg_parser_stream_get_current_format(pin->wg_stream, &format);
+    if (format.major_type == WG_MAJOR_TYPE_VIDEO || format.major_type == WG_MAJOR_TYPE_AUDIO)
+        return amt_to_wg_format(mt, &format) ? S_OK : S_FALSE;
     if (!amt_from_wg_format(&pad_mt, &format, false))
         return E_OUTOFMEMORY;
     hr = compare_media_types(mt, &pad_mt) ? S_OK : S_FALSE;
@@ -2435,9 +2307,11 @@
 {
     struct wg_format format;
 
+    wg_parser_stream_get_current_format(pin->wg_stream, &format);
+    if (format.major_type == WG_MAJOR_TYPE_VIDEO || format.major_type == WG_MAJOR_TYPE_AUDIO)
+        return get_raw_media_type(&format, index, mt);
     if (index > 0)
         return VFW_S_NO_MORE_ITEMS;
-    wg_parser_stream_get_current_format(pin->wg_stream, &format);
     if (!amt_from_wg_format(mt, &format, false))
         return E_OUTOFMEMORY;
     return S_OK;
@@ -2504,25 +2378,12 @@
     HRESULT hr;
 
     wg_parser_stream_get_current_format(pin->wg_stream, &format);
-    if (wg_video_format_is_rgb(format.u.video.format))
-        format.u.video.height = -format.u.video.height;
+    if (format.major_type == WG_MAJOR_TYPE_VIDEO || format.major_type == WG_MAJOR_TYPE_AUDIO)
+        return amt_to_wg_format(mt, &format) ? S_OK : S_FALSE;
     if (!amt_from_wg_format(&pad_mt, &format, false))
         return E_OUTOFMEMORY;
     hr = compare_media_types(mt, &pad_mt) ? S_OK : S_FALSE;
     FreeMediaType(&pad_mt);
-
-    /* For 32 bpp AVI files, also accept ARGB32 format. */
-    if (hr == S_FALSE
-        && format.major_type == WG_MAJOR_TYPE_VIDEO
-        && format.u.video.format == WG_VIDEO_FORMAT_BGRx)
-    {
-        format.u.video.format = WG_VIDEO_FORMAT_BGRA;
-        if (!amt_from_wg_format(&pad_mt, &format, false))
-            return E_OUTOFMEMORY;
-        hr = compare_media_types(mt, &pad_mt) ? S_OK : S_FALSE;
-        FreeMediaType(&pad_mt);
-    }
-
     return hr;
 }
 
@@ -2531,22 +2392,11 @@
 {
     struct wg_format format;
 
-    if (index > 1)
+    wg_parser_stream_get_current_format(pin->wg_stream, &format);
+    if (format.major_type == WG_MAJOR_TYPE_VIDEO || format.major_type == WG_MAJOR_TYPE_AUDIO)
+        return get_raw_media_type(&format, index, mt);
+    if (index > 0)
         return VFW_S_NO_MORE_ITEMS;
-
-    wg_parser_stream_get_current_format(pin->wg_stream, &format);
-    if (format.major_type == WG_MAJOR_TYPE_VIDEO
-        && format.u.video.format == WG_VIDEO_FORMAT_BGRx)
-    {
-        /* Offer ARGB32 format as the first candidate for 32 bpp AVI files. */
-        if (index == 0)
-            format.u.video.format = WG_VIDEO_FORMAT_BGRA;
-    }
-    else if (index > 0)
-        return VFW_S_NO_MORE_ITEMS;
-
-    if (wg_video_format_is_rgb(format.u.video.format))
-        format.u.video.height = -format.u.video.height;
     if (!amt_from_wg_format(mt, &format, false))
         return E_OUTOFMEMORY;
     return S_OK;
@@ -2671,6 +2521,8 @@
     HRESULT hr;
 
     wg_parser_stream_get_current_format(pin->wg_stream, &format);
+    if (format.major_type == WG_MAJOR_TYPE_VIDEO || format.major_type == WG_MAJOR_TYPE_AUDIO)
+        return amt_to_wg_format(mt, &format) ? S_OK : S_FALSE;
     if (!amt_from_wg_format(&pad_mt, &format, false))
         return E_OUTOFMEMORY;
     hr = compare_media_types(mt, &pad_mt) ? S_OK : S_FALSE;
@@ -2683,9 +2535,11 @@
 {
     struct wg_format format;
 
+    wg_parser_stream_get_current_format(pin->wg_stream, &format);
+    if (format.major_type == WG_MAJOR_TYPE_VIDEO || format.major_type == WG_MAJOR_TYPE_AUDIO)
+        return get_raw_media_type(&format, index, mt);
     if (index > 0)
         return VFW_S_NO_MORE_ITEMS;
-    wg_parser_stream_get_current_format(pin->wg_stream, &format);
     if (!amt_from_wg_format(mt, &format, false))
         return E_OUTOFMEMORY;
     return S_OK;
@@ -2729,7 +2583,7 @@
     object->init_gst = mpeg_splitter_filter_init_gst;
     object->source_query_accept = mpeg_splitter_source_query_accept;
     object->source_get_media_type = mpeg_splitter_source_get_media_type;
-    object->enum_sink_first = true;
+    object->enum_sink_first = TRUE;
 
     TRACE("Created MPEG-1 splitter %p.\n", object);
     *out = &object->filter.IUnknown_inner;
