diff --git a/dlls/mf/session.c b/dlls/mf/session.c
--- a/dlls/mf/session.c
+++ b/dlls/mf/session.c
@@ -112,9 +112,6 @@
     /* STARTED -> PAUSED transition */
     COMMAND_STATE_PAUSING_SINKS,      /* -> COMMAND_STATE_PAUSING_SOURCES */
     COMMAND_STATE_PAUSING_SOURCES,    /* -> SESSION_STATE_PAUSED */
-    /* STARTED -> STOPPED transition when presentation ends */
-    COMMAND_STATE_ENDING_STREAMS,     /* -> COMMAND_STATE_ENDING_SINKS */
-    COMMAND_STATE_ENDING_SINKS,       /* -> SESSION_STATE_STOPPED */
     /* STARTED | PAUSED -> STOPPED transition */
     COMMAND_STATE_STOPPING_SINKS,     /* -> COMMAND_STATE_STOPPING_SOURCES */
     COMMAND_STATE_STOPPING_SOURCES,   /* -> SESSION_STATE_STOPPED */
@@ -175,6 +172,7 @@
     struct list samples;
     unsigned int requests;
     unsigned int min_buffer_size;
+    IMFSample *allocated_sample;
     BOOL draining;
 };
 
@@ -237,6 +235,8 @@
     SESSION_FLAG_SOURCE_SHUTDOWN = 0x10,
     SESSION_FLAG_PENDING_RATE_CHANGE = 0x20,
     SESSION_FLAG_RESTARTING = 0x40,
+    SESSION_FLAG_SINKS_SUBSCRIBED = 0x80,
+    SESSION_FLAG_PRESENTATION_ENDING = 0x100,
 };
 
 struct media_session
@@ -484,6 +484,7 @@
     EnterCriticalSection(&session->cs);
     if (SUCCEEDED(hr = session_is_shut_down(session)))
     {
+        session->presentation.flags &= ~SESSION_FLAG_PRESENTATION_ENDING;
         if (list_empty(&session->commands) && session->command_state == COMMAND_STATE_COMPLETE)
         {
             hr = MFPutWorkItem(MFASYNC_CALLBACK_QUEUE_STANDARD, &session->commands_callback, &op->IUnknown_iface);
@@ -533,7 +534,7 @@
     IMFMediaEvent *event;
     PROPVARIANT param;
 
-    if (topo_status == MF_TOPOSTATUS_INVALID)
+    if (topo_status == MF_TOPOSTATUS_INVALID || status == MF_E_SHUTDOWN)
         return;
 
     if (list_empty(&session->topologies))
@@ -775,6 +776,12 @@
 {
     IMFMediaEvent *event;
 
+    if (stream->allocated_sample)
+    {
+        IMFSample_Release(stream->allocated_sample);
+        stream->allocated_sample = NULL;
+    }
+
     while (SUCCEEDED(transform_stream_pop_event(stream, &event)))
         IMFMediaEvent_Release(event);
 }
@@ -853,7 +860,7 @@
                         WARN("Failed to shut down activation object for the sink, hr %#lx.\n", hr);
                     IMFActivate_Release(activate);
                 }
-                else if (SUCCEEDED(topology_node_get_object(node, &IID_IMFStreamSink, (void **)&stream_sink)))
+                if (SUCCEEDED(topology_node_get_object(node, &IID_IMFStreamSink, (void **)&stream_sink)))
                 {
                     if (SUCCEEDED(IMFStreamSink_GetMediaSink(stream_sink, &sink)))
                     {
@@ -1016,6 +1023,30 @@
     return hr;
 }
 
+static HRESULT session_subscribe_sinks(struct media_session *session)
+{
+    struct topo_node *node;
+    HRESULT hr = S_OK;
+
+    if (session->presentation.flags & SESSION_FLAG_SINKS_SUBSCRIBED)
+        return hr;
+
+    LIST_FOR_EACH_ENTRY(node, &session->presentation.nodes, struct topo_node, entry)
+    {
+        if (node->type != MF_TOPOLOGY_OUTPUT_NODE)
+            continue;
+
+        if (FAILED(hr = IMFStreamSink_BeginGetEvent(node->object.sink_stream, &session->events_callback,
+                        node->object.object)))
+        {
+            WARN("Failed to subscribe to stream sink events, hr %#lx.\n", hr);
+        }
+    }
+
+    session->presentation.flags |= SESSION_FLAG_SINKS_SUBSCRIBED;
+    return hr;
+}
+
 static void session_flush_transforms(struct media_session *session)
 {
     struct topo_node *node;
@@ -1055,47 +1086,59 @@
 static void session_flush_nodes(struct media_session *session)
 {
     struct topo_node *node;
-    UINT i;
 
     LIST_FOR_EACH_ENTRY(node, &session->presentation.nodes, struct topo_node, entry)
     {
         if (node->type == MF_TOPOLOGY_OUTPUT_NODE)
             IMFStreamSink_Flush(node->object.sink_stream);
         else if (node->type == MF_TOPOLOGY_TRANSFORM_NODE)
-        {
-            for (i = 0; i < node->u.transform.output_count; ++i)
-                node->u.transform.outputs[i].requests = 0;
-
             IMFTransform_ProcessMessage(node->object.transform, MFT_MESSAGE_COMMAND_FLUSH, 0);
-        }
     }
 }
 
 static void session_handle_source_shutdown(struct media_session *session);
 
+static void session_reset_transforms(struct media_session *session, BOOL drop)
+{
+    struct topo_node *topo_node;
+    UINT i;
+
+    LIST_FOR_EACH_ENTRY(topo_node, &session->presentation.nodes, struct topo_node, entry)
+    {
+        if (topo_node->type != MF_TOPOLOGY_TRANSFORM_NODE)
+            continue;
+
+        for (i = 0; i < topo_node->u.transform.input_count; i++)
+        {
+            struct transform_stream *stream = &topo_node->u.transform.inputs[i];
+            stream->draining = FALSE;
+            if (drop)
+                transform_stream_drop_events(stream);
+        }
+
+        if (!drop)
+            continue;
+
+        for (i = 0; i < topo_node->u.transform.output_count; ++i)
+        {
+            struct transform_stream *stream = &topo_node->u.transform.outputs[i];
+            transform_stream_drop_events(stream);
+            stream->requests = 0;
+        }
+    }
+}
+
 static void session_start(struct media_session *session, const GUID *time_format, const PROPVARIANT *start_position)
 {
     struct media_source *source;
-    struct topo_node *topo_node;
-    BOOL keep_position;
     MFTIME duration;
     HRESULT hr;
-    UINT i;
-
-    keep_position = IsEqualGUID(time_format, &GUID_NULL) && start_position->vt == VT_EMPTY;
-
-    /* No position change - nothing to do. */
-    if (session->state == SESSION_STATE_STARTED && keep_position)
-    {
-        session_command_complete_with_event(session, MESessionStarted, S_OK, NULL);
-        return;
-    }
 
     switch (session->state)
     {
         case SESSION_STATE_PAUSED:
         case SESSION_STATE_STARTED:
-            if (!keep_position)
+            if (!IsEqualGUID(time_format, &GUID_NULL) || start_position->vt != VT_EMPTY)
             {
                 session->command_state = COMMAND_STATE_RESTARTING_SOURCES;
 
@@ -1120,7 +1163,8 @@
                         WARN("Failed to stop media source %p, hr %#lx.\n", source->source, hr);
                         if (hr == MF_E_SHUTDOWN)
                             return session_handle_source_shutdown(session);
-                        return session_command_complete_with_event(session, MESessionStarted, hr, NULL);
+                        session_command_complete_with_event(session, MESessionStarted, hr, NULL);
+                        return;
                     }
                 }
 
@@ -1130,6 +1174,9 @@
 
                 break;
             }
+            else if (session->state == SESSION_STATE_STARTED)
+                return session_command_complete_with_event(session, MESessionStarted, S_OK, NULL);
+
             /* fallthrough; we're resuming from the current position */
         case SESSION_STATE_STOPPED:
             /* Start request with no current topology. */
@@ -1151,6 +1198,8 @@
                     return session_handle_source_shutdown(session);
                 return session_command_complete_with_event(session, MESessionStarted, hr, NULL);
             }
+
+            session_reset_transforms(session, FALSE);
 
             LIST_FOR_EACH_ENTRY(source, &session->presentation.sources, struct media_source, entry)
             {
@@ -1163,17 +1212,6 @@
                 }
             }
 
-            LIST_FOR_EACH_ENTRY(topo_node, &session->presentation.nodes, struct topo_node, entry)
-            {
-                if (topo_node->type == MF_TOPOLOGY_TRANSFORM_NODE)
-                {
-                    for (i = 0; i < topo_node->u.transform.input_count; i++)
-                    {
-                        struct transform_stream *stream = &topo_node->u.transform.inputs[i];
-                        stream->draining = FALSE;
-                    }
-                }
-            }
             break;
 
         case SESSION_STATE_CLOSED:
@@ -1283,6 +1321,7 @@
 {
     IMFMediaEvent *event;
 
+    session->presentation.flags &= ~SESSION_FLAG_PRESENTATION_ENDING;
     session->state = SESSION_STATE_STOPPED;
 
     if (SUCCEEDED(MFCreateMediaEvent(event_type, &GUID_NULL, status, NULL, &event)))
@@ -1497,18 +1536,6 @@
 
         if (FAILED(hr))
             WARN("Failed to set time source, hr %#lx.\n", hr);
-
-        LIST_FOR_EACH_ENTRY(node, &session->presentation.nodes, struct topo_node, entry)
-        {
-            if (node->type != MF_TOPOLOGY_OUTPUT_NODE)
-                continue;
-
-            if (FAILED(hr = IMFStreamSink_BeginGetEvent(node->object.sink_stream, &session->events_callback,
-                    node->object.object)))
-            {
-                WARN("Failed to subscribe to stream sink events, hr %#lx.\n", hr);
-            }
-        }
 
         /* Set clock for all topology nodes. */
         LIST_FOR_EACH_ENTRY(source, &session->presentation.sources, struct media_source, entry)
@@ -1880,7 +1907,7 @@
                         &IID_IMFVideoSampleAllocator, (void **)&topo_node->u.sink.allocator)))
                     {
                         if (FAILED(hr = IMFVideoSampleAllocator_InitializeSampleAllocator(topo_node->u.sink.allocator,
-                                2, media_type)))
+                                4, media_type)))
                         {
                             WARN("Failed to initialize sample allocator for the stream, hr %#lx.\n", hr);
                         }
@@ -2006,7 +2033,8 @@
         return hr;
     }
 
-    session_collect_nodes(session);
+    if (FAILED(hr = session_collect_nodes(session)))
+        return hr;
 
     LIST_FOR_EACH_ENTRY(node, &session->presentation.nodes, struct topo_node, entry)
     {
@@ -2084,10 +2112,14 @@
         {
             hr = session_bind_output_nodes(topology);
 
+            IMFTopology_SetUINT32(topology, &MF_TOPOLOGY_ENUMERATE_SOURCE_TYPES, TRUE);
+
             if (SUCCEEDED(hr))
                 hr = IMFTopoLoader_Load(session->topo_loader, topology, &resolved_topology, NULL /* FIXME? */);
             if (SUCCEEDED(hr))
                 hr = session_init_media_types(resolved_topology);
+            else
+                WARN("failed to load topology %p, hr %#lx.\n", topology, hr);
 
             if (SUCCEEDED(hr))
             {
@@ -2117,7 +2149,7 @@
     session_raise_topology_set(session, topology, hr);
 
     /* With no current topology set it right away, otherwise queue. */
-    if (topology)
+    if (SUCCEEDED(hr) && topology)
     {
         struct queued_topology *queued_topology;
 
@@ -2732,10 +2764,7 @@
         hr = session_get_audio_render_service(session, service, riid, obj);
     }
     else
-    {
         FIXME("Unsupported service %s.\n", debugstr_guid(service));
-        hr = E_NOINTERFACE;
-    }
 
     LeaveCriticalSection(&session->cs);
 
@@ -2944,8 +2973,6 @@
             session->state = SESSION_STATE_STOPPED;
             session_command_complete_with_event(session, MESessionPaused, MF_E_SHUTDOWN, NULL);
             break;
-        case COMMAND_STATE_ENDING_STREAMS:
-        case COMMAND_STATE_ENDING_SINKS:
         case COMMAND_STATE_STOPPING_SINKS:
         case COMMAND_STATE_STOPPING_SOURCES:
             session_clear_presentation(session);
@@ -3212,6 +3239,7 @@
                 }
             }
 
+            session_subscribe_sinks(session);
             session_set_presentation_clock(session);
 
             if ((session->presentation.flags & SESSION_FLAG_NEEDS_PREROLL) && session_is_output_nodes_state(session, OBJ_STATE_STOPPED))
@@ -3277,10 +3305,12 @@
             if (!session_is_source_nodes_state(session, OBJ_STATE_STOPPED))
                 break;
 
+            session->presentation.flags &= ~SESSION_FLAG_PRESENTATION_ENDING;
+            session->presentation.flags |= SESSION_FLAG_RESTARTING;
+            session_reset_transforms(session, TRUE);
+            session_flush_transforms(session);
             session->state = SESSION_STATE_STOPPED;
             session->command_state = COMMAND_STATE_STARTING_SOURCES;
-            session->presentation.flags |= SESSION_FLAG_RESTARTING;
-            session_flush_transforms(session);
 
             /* Start sources */
             LIST_FOR_EACH_ENTRY(source, &session->presentation.sources, struct media_source, entry)
@@ -3323,8 +3353,6 @@
         case COMMAND_STATE_STARTING_SINKS:
         case COMMAND_STATE_PAUSING_SINKS:
         case COMMAND_STATE_STOPPING_SINKS:
-        case COMMAND_STATE_ENDING_STREAMS:
-        case COMMAND_STATE_ENDING_SINKS:
         case COMMAND_STATE_CLOSING_SINKS:
         case COMMAND_STATE_FINALIZING_SINKS:
             WARN("Ignoring source state change in command state %#x\n", session->command_state);
@@ -3377,15 +3405,6 @@
                 session_set_paused(session, hr);
 
             break;
-        case COMMAND_STATE_ENDING_SINKS:
-            if (!session_is_output_nodes_state(session, OBJ_STATE_STOPPED))
-                break;
-
-            LIST_FOR_EACH_ENTRY(source, &session->presentation.sources, struct media_source, entry)
-                IMFMediaSource_Stop(source->source);
-
-            session_set_stopped(session, MESessionEnded, S_OK);
-            break;
         case COMMAND_STATE_STOPPING_SINKS:
             if (!session_is_output_nodes_state(session, OBJ_STATE_STOPPED))
                 break;
@@ -3398,6 +3417,8 @@
 
             if (FAILED(hr))
                 session_set_stopped(session, MESessionStopped, hr);
+            else if (session->presentation.flags & SESSION_FLAG_PRESENTATION_ENDING)
+                session_set_stopped(session, MESessionEnded, S_OK);
             break;
         case COMMAND_STATE_CLOSING_SINKS:
             if (!session_is_output_nodes_state(session, OBJ_STATE_STOPPED))
@@ -3417,7 +3438,6 @@
         case COMMAND_STATE_RESTARTING_SOURCES:
         case COMMAND_STATE_STARTING_SOURCES:
         case COMMAND_STATE_PAUSING_SOURCES:
-        case COMMAND_STATE_ENDING_STREAMS:
         case COMMAND_STATE_STOPPING_SOURCES:
         case COMMAND_STATE_CLOSING_SOURCES:
         case COMMAND_STATE_FINALIZING_SINKS:
@@ -3429,11 +3449,21 @@
 static HRESULT transform_get_external_output_sample(const struct media_session *session, struct topo_node *transform,
         DWORD output, const MFT_OUTPUT_STREAM_INFO *stream_info, IMFSample **sample)
 {
+    struct transform_stream *stream = &transform->u.transform.outputs[output];
+    DWORD buffer_size, sample_size, input;
     IMFMediaBuffer *buffer = NULL;
     struct topo_node *topo_node;
-    unsigned int buffer_size;
-    DWORD input;
     HRESULT hr;
+
+    buffer_size = max(stream_info->cbSize, stream->min_buffer_size);
+    if ((*sample = stream->allocated_sample))
+    {
+        stream->allocated_sample = NULL;
+        if (SUCCEEDED(IMFSample_GetTotalLength(*sample, &sample_size)) && sample_size >= buffer_size)
+            return S_OK;
+        IMFSample_Release(*sample);
+        *sample = NULL;
+    }
 
     if (!(topo_node = session_get_topo_node_output(session, transform, output, &input)))
     {
@@ -3447,8 +3477,6 @@
     }
     else
     {
-        buffer_size = max(stream_info->cbSize, transform->u.transform.outputs[output].min_buffer_size);
-
         hr = MFCreateAlignedMemoryBuffer(buffer_size, stream_info->cbAlignment, &buffer);
         if (SUCCEEDED(hr))
             hr = MFCreateSample(sample);
@@ -3666,6 +3694,11 @@
             if (FAILED(hr = transform_stream_push_sample(stream, buffers[i].pSample)))
                 WARN("Failed to queue output sample, hr %#lx\n", hr);
         }
+        else if (hr == MF_E_TRANSFORM_NEED_MORE_INPUT && !stream->allocated_sample)
+        {
+            stream->allocated_sample = buffers[i].pSample;
+            buffers[i].pSample = NULL;
+        }
     }
 
 done:
@@ -3706,13 +3739,32 @@
     struct transform_stream *stream = &topo_node->u.transform.inputs[input];
     UINT id = transform_node_get_stream_id(topo_node, FALSE, input);
     IMFTransform *transform = topo_node->object.transform;
+    MFTIME pts, clocktime;
+    IMFMediaType *type;
+    GUID major_type;
+    const char *sgi;
     HRESULT hr;
 
     if (sample)
     {
-        hr = IMFTransform_ProcessInput(transform, id, sample, 0);
-        if (hr == MF_E_NOTACCEPTING)
-            hr = transform_stream_push_sample(stream, sample);
+        /* HACK: VRChat (438100) requires audio packets that are late to be dropped prior to being passed to an MFT */
+        if((sgi = getenv("SteamGameId")) && !strcmp(sgi, "438100") &&
+            SUCCEEDED(IMFTransform_GetOutputCurrentType(transform, id, &type)) &&
+            SUCCEEDED(IMFMediaType_GetMajorType(type, &major_type)) &&
+            IsEqualGUID(&major_type, &MFMediaType_Audio) &&
+            SUCCEEDED(IMFSample_GetSampleTime(sample, &pts)) && pts != MINLONGLONG &&
+            SUCCEEDED(IMFPresentationClock_GetTime(session->clock, &clocktime)) &&
+            clocktime > pts)
+        {
+            hr = S_OK;
+            FIXME("dropping audio sample clocktime %I64d pts %I64d jitter: %I64d\n", clocktime, pts, clocktime - pts);
+        }
+        else
+        {
+            hr = IMFTransform_ProcessInput(transform, id, sample, 0);
+            if (hr == MF_E_NOTACCEPTING)
+                hr = transform_stream_push_sample(stream, sample);
+        }
     }
     else
     {
@@ -4162,7 +4214,7 @@
     {
         if (session_nodes_is_mask_set(session, MF_TOPOLOGY_MAX, SOURCE_FLAG_END_OF_PRESENTATION))
         {
-            session->command_state = COMMAND_STATE_ENDING_STREAMS;
+            session->presentation.flags |= SESSION_FLAG_PRESENTATION_ENDING;
             IMFMediaEventQueue_QueueEventParamVar(session->event_queue, MEEndOfPresentation, &GUID_NULL, S_OK, NULL);
         }
     }
@@ -4215,7 +4267,7 @@
 
     node->flags |= TOPO_NODE_END_OF_STREAM;
 
-    if (session->command_state == COMMAND_STATE_ENDING_STREAMS &&
+    if (session->presentation.flags & SESSION_FLAG_PRESENTATION_ENDING &&
             session_nodes_is_mask_set(session, MF_TOPOLOGY_OUTPUT_NODE, TOPO_NODE_END_OF_STREAM))
     {
         session_set_topo_status(session, S_OK, MF_TOPOSTATUS_ENDED);
@@ -4223,7 +4275,7 @@
 
         IMFPresentationClock_GetTime(session->clock, &session->presentation.clock_stop_time);
         if (SUCCEEDED(hr = IMFPresentationClock_Stop(session->clock)))
-            session->command_state = COMMAND_STATE_ENDING_SINKS;
+            session->command_state = COMMAND_STATE_STOPPING_SINKS;
         else
             session_set_stopped(session, MESessionEnded, hr);
     }
@@ -4757,6 +4809,17 @@
     return IMFMediaSession_Release(&session->IMFMediaSession_iface);
 }
 
+static BOOL allow_rate_zero(void)
+{
+    static int allow = -1;
+    if (allow == -1)
+    {
+        const char *sgi = getenv("SteamGameId");
+        allow = sgi && !strcmp(sgi, "2111630") /* JR East Train Simulator */ ;
+    }
+    return allow;
+}
+
 static HRESULT WINAPI session_rate_control_SetRate(IMFRateControl *iface, BOOL thin, float rate)
 {
     struct media_session *session = impl_session_from_IMFRateControl(iface);
@@ -4764,6 +4827,13 @@
     HRESULT hr;
 
     TRACE("%p, %d, %f.\n", iface, thin, rate);
+
+    if (!rate && !allow_rate_zero())
+    {
+        /* The Anacrusis fails to play its video if we succeed here */
+        ERR("Scrubbing not implemented!\n");
+        return E_NOTIMPL;
+    }
 
     if (FAILED(hr = create_session_op(SESSION_CMD_SET_RATE, &op)))
         return hr;
@@ -4932,6 +5002,8 @@
         goto failed;
     }
 
+    session_clear_presentation(object);
+
     *session = &object->IMFMediaSession_iface;
 
     return S_OK;
