diff --git a/dlls/cryptui/main.c b/dlls/cryptui/main.c
--- a/dlls/cryptui/main.c
+++ b/dlls/cryptui/main.c
@@ -42,6 +42,8 @@
 
 static HINSTANCE hInstance;
 
+static const WCHAR empty[] = {0};
+
 BOOL WINAPI DllMain(HINSTANCE hinstDLL, DWORD fdwReason, LPVOID lpvReserved)
 {
     TRACE("(0x%p, %ld, %p)\n", hinstDLL, fdwReason, lpvReserved);
@@ -172,8 +174,10 @@
 
 static LPSTR get_cert_mgr_usages(void)
 {
-    static const WCHAR keyName[] =
-      L"Software\\Microsoft\\Cryptography\\UI\\Certmgr\\Purpose";
+    static const WCHAR keyName[] = { 'S','o','f','t','w','a','r','e','\\','M',
+     'i','c','r','o','s','o','f','t','\\','C','r','y','p','t','o','g','r','a',
+     'p','h','y','\\','U','I','\\','C','e','r','t','m','g','r','\\','P','u',
+     'r','p','o','s','e',0 };
     LPSTR str = NULL;
     HKEY key;
 
@@ -433,6 +437,15 @@
      (LPARAM)cert_mgr_sort_by_subject);
 }
 
+static const WCHAR my[] = { 'M','y',0 };
+static const WCHAR addressBook[] = {
+ 'A','d','d','r','e','s','s','B','o','o','k',0 };
+static const WCHAR ca[] = { 'C','A',0 };
+static const WCHAR root[] = { 'R','o','o','t',0 };
+static const WCHAR trustedPublisher[] = {
+ 'T','r','u','s','t','e','d','P','u','b','l','i','s','h','e','r',0 };
+static const WCHAR disallowed[] = { 'D','i','s','a','l','l','o','w','e','d',0 };
+
 struct CertMgrStoreInfo
 {
     LPCWSTR name;
@@ -441,21 +454,21 @@
 };
 
 static const struct CertMgrStoreInfo defaultStoreList[] = {
- { L"My", IDS_WARN_REMOVE_MY, IDS_WARN_REMOVE_PLURAL_MY },
- { L"AddressBook", IDS_WARN_REMOVE_ADDRESSBOOK,
+ { my, IDS_WARN_REMOVE_MY, IDS_WARN_REMOVE_PLURAL_MY },
+ { addressBook, IDS_WARN_REMOVE_ADDRESSBOOK,
    IDS_WARN_REMOVE_PLURAL_ADDRESSBOOK },
- { L"CA", IDS_WARN_REMOVE_CA, IDS_WARN_REMOVE_PLURAL_CA },
- { L"Root", IDS_WARN_REMOVE_ROOT, IDS_WARN_REMOVE_PLURAL_ROOT },
- { L"TrustedPublisher", IDS_WARN_REMOVE_TRUSTEDPUBLISHER,
+ { ca, IDS_WARN_REMOVE_CA, IDS_WARN_REMOVE_PLURAL_CA },
+ { root, IDS_WARN_REMOVE_ROOT, IDS_WARN_REMOVE_PLURAL_ROOT },
+ { trustedPublisher, IDS_WARN_REMOVE_TRUSTEDPUBLISHER,
    IDS_WARN_REMOVE_PLURAL_TRUSTEDPUBLISHER },
- { L"Disallowed", IDS_WARN_REMOVE_DEFAULT },
+ { disallowed, IDS_WARN_REMOVE_DEFAULT },
 };
 
 static const struct CertMgrStoreInfo publisherStoreList[] = {
- { L"Root", IDS_WARN_REMOVE_ROOT, IDS_WARN_REMOVE_PLURAL_ROOT },
- { L"TrustedPublisher", IDS_WARN_REMOVE_TRUSTEDPUBLISHER,
+ { root, IDS_WARN_REMOVE_ROOT, IDS_WARN_REMOVE_PLURAL_ROOT },
+ { trustedPublisher, IDS_WARN_REMOVE_TRUSTEDPUBLISHER,
    IDS_WARN_REMOVE_PLURAL_TRUSTEDPUBLISHER },
- { L"Disallowed", IDS_WARN_REMOVE_PLURAL_DEFAULT },
+ { disallowed, IDS_WARN_REMOVE_PLURAL_DEFAULT },
 };
 
 struct CertMgrData
@@ -636,8 +649,10 @@
 
 static void save_cert_mgr_usages(HWND hwnd)
 {
-    static const WCHAR keyName[] =
-      L"Software\\Microsoft\\Cryptography\\UI\\Certmgr\\Purpose";
+    static const WCHAR keyName[] = { 'S','o','f','t','w','a','r','e','\\','M',
+     'i','c','r','o','s','o','f','t','\\','C','r','y','p','t','o','g','r','a',
+     'p','h','y','\\','U','I','\\','C','e','r','t','m','g','r','\\','P','u',
+     'r','p','o','s','e',0 };
     HKEY key;
     HWND lv = GetDlgItem(hwnd, IDC_CERTIFICATE_USAGES);
     int purposes = SendMessageW(lv, LVM_GETITEMCOUNT, 0, 0), i;
@@ -778,7 +793,8 @@
     EnableWindow(GetDlgItem(hwnd, IDC_MGR_EXPORT), FALSE);
     EnableWindow(GetDlgItem(hwnd, IDC_MGR_REMOVE), FALSE);
     EnableWindow(GetDlgItem(hwnd, IDC_MGR_VIEW), FALSE);
-    SendMessageW(GetDlgItem(hwnd, IDC_MGR_PURPOSES), WM_SETTEXT, 0, (LPARAM)L"");
+    SendMessageW(GetDlgItem(hwnd, IDC_MGR_PURPOSES), WM_SETTEXT, 0,
+     (LPARAM)empty);
     refresh_store_certs(hwnd);
 }
 
@@ -851,6 +867,7 @@
     {
         if (usage->cUsageIdentifier)
         {
+            static const WCHAR commaSpace[] = { ',',' ',0 };
             DWORD i, len = 1;
             LPWSTR ptr;
 
@@ -866,7 +883,7 @@
                 else
                     len += strlen(usage->rgpszUsageIdentifier[i]);
                 if (i < usage->cUsageIdentifier - 1)
-                    len += strlen(", ");
+                    len += lstrlenW(commaSpace);
             }
             *str = malloc(len * sizeof(WCHAR));
             if (*str)
@@ -893,8 +910,8 @@
                     }
                     if (i < usage->cUsageIdentifier - 1)
                     {
-                        lstrcpyW(ptr, L", ");
-                        ptr += strlen(", ");
+                        lstrcpyW(ptr, commaSpace);
+                        ptr += lstrlenW(commaSpace);
                     }
                 }
                 *ptr = 0;
@@ -970,9 +987,7 @@
                     PCCERT_CONTEXT cert = cert_mgr_index_to_cert(hwnd,
                      selection);
 
-                    /* CertDeleteCertificateFromStore() releases certificate, duplicate it to keep context stored
-                     * in the list valid. */
-                    CertDeleteCertificateFromStore(CertDuplicateCertificateContext(cert));
+                    CertDeleteCertificateFromStore(cert);
                 }
             } while (selection >= 0);
             cert_mgr_clear_cert_selection(hwnd);
@@ -1145,7 +1160,8 @@
             if (numSelected == 1)
                 cert_mgr_show_cert_usages(hwnd, nm->iItem);
             else
-                SendMessageW(GetDlgItem(hwnd, IDC_MGR_PURPOSES), WM_SETTEXT, 0, (LPARAM)L"");
+                SendMessageW(GetDlgItem(hwnd, IDC_MGR_PURPOSES), WM_SETTEXT, 0,
+                 (LPARAM)empty);
             break;
         }
         case NM_DBLCLK:
@@ -1449,18 +1465,17 @@
     }
 }
 
-static WCHAR selected_store_title[MAX_STRING_LEN];
-
 static HCERTSTORE selected_item_to_store(HWND tree, HTREEITEM hItem)
 {
+    WCHAR buf[MAX_STRING_LEN];
     TVITEMW item;
     HCERTSTORE store;
 
     memset(&item, 0, sizeof(item));
     item.mask = TVIF_HANDLE | TVIF_PARAM | TVIF_TEXT;
     item.hItem = hItem;
-    item.cchTextMax = ARRAY_SIZE(selected_store_title);
-    item.pszText = selected_store_title;
+    item.cchTextMax = ARRAY_SIZE(buf);
+    item.pszText = buf;
     SendMessageW(tree, TVM_GETITEMW, 0, (LPARAM)&item);
     if (item.lParam)
     {
@@ -1474,7 +1489,7 @@
     else
     {
         /* It's implicitly a system store */
-        store = CertOpenSystemStoreW(0, selected_store_title);
+        store = CertOpenSystemStoreW(0, buf);
     }
     return store;
 }
@@ -1504,7 +1519,7 @@
             SendMessageW(GetDlgItem(hwnd, IDC_STORE_TEXT), WM_SETTEXT, 0,
              (LPARAM)selectInfo->info->pwszText);
         if (!(selectInfo->info->dwFlags & CRYPTUI_ENABLE_SHOW_PHYSICAL_STORE))
-            ShowWindow(GetDlgItem(hwnd, IDC_SHOW_PHYSICAL_STORES), SW_HIDE);
+            ShowWindow(GetDlgItem(hwnd, IDC_SHOW_PHYSICAL_STORES), FALSE);
         enumerate_stores(hwnd, selectInfo->info->pEnumData);
         break;
     }
@@ -2468,26 +2483,16 @@
         page = (PROPSHEETPAGEW *)lp;
         pCertViewInfo = (PCCRYPTUI_VIEWCERTIFICATE_STRUCTW)page->lParam;
         if (pCertViewInfo->dwFlags & CRYPTUI_DISABLE_ADDTOSTORE)
-            ShowWindow(GetDlgItem(hwnd, IDC_ADDTOSTORE), SW_HIDE);
+            ShowWindow(GetDlgItem(hwnd, IDC_ADDTOSTORE), FALSE);
         EnableWindow(GetDlgItem(hwnd, IDC_ISSUERSTATEMENT), FALSE);
         set_general_info(hwnd, pCertViewInfo);
-        SetWindowLongPtrW(hwnd, GWLP_USERDATA, (LONG_PTR)pCertViewInfo->pCertContext);
         break;
     case WM_COMMAND:
         switch (wp)
         {
         case IDC_ADDTOSTORE:
-        {
-            CRYPTUI_WIZ_IMPORT_SRC_INFO info;
-
-            info.dwSize = sizeof(info);
-            info.dwSubjectChoice = CRYPTUI_WIZ_IMPORT_SUBJECT_CERT_CONTEXT;
-            info.pCertContext = (PCCERT_CONTEXT)GetWindowLongPtrW(hwnd, GWLP_USERDATA);
-            info.dwFlags = 0;
-            info.pwszPassword = L"";
-            CryptUIWizImport(0, hwnd, NULL, &info, NULL);
-            break;
-        }
+            CryptUIWizImport(0, hwnd, NULL, NULL, NULL);
+            break;
         case IDC_ISSUERSTATEMENT:
         {
             struct IssuerStatement *issuerStatement =
@@ -2553,10 +2558,11 @@
 
 static WCHAR *field_format_version(PCCERT_CONTEXT cert)
 {
+    static const WCHAR fmt[] = { 'V','%','d',0 };
     WCHAR *buf = malloc(12 * sizeof(WCHAR));
 
     if (buf)
-        swprintf(buf, 12, L"V%d", cert->pCertInfo->dwVersion);
+        swprintf(buf, 12, fmt, cert->pCertInfo->dwVersion);
     return buf;
 }
 
@@ -2566,11 +2572,12 @@
 
     if (buf)
     {
+        static const WCHAR fmt[] = { '%','0','2','x',' ',0 };
         DWORD i;
         WCHAR *ptr;
 
         for (i = 0, ptr = buf; i < cb; i++, ptr += 3)
-            swprintf(ptr, 4, L"%02x ", ((BYTE *)pb)[i]);
+            swprintf(ptr, 4, fmt, ((BYTE *)pb)[i]);
     }
     return buf;
 }
@@ -2851,6 +2858,7 @@
         str = malloc((lines * 8 * 4 + lines * 3 + 1) * sizeof(WCHAR));
         if (str)
         {
+            static const WCHAR fmt[] = { '%','0','2','x',' ',0 };
             DWORD i, j;
             WCHAR *ptr;
 
@@ -2858,7 +2866,7 @@
             {
                 /* Output as hex bytes first */
                 for (j = i; j < min(i + 8, ext->Value.cbData); j++, ptr += 3)
-                    swprintf(ptr, 4, L"%02x ", ext->Value.pbData[j]);
+                    swprintf(ptr, 4, fmt, ext->Value.pbData[j]);
                 /* Pad the hex output with spaces for alignment */
                 if (j == ext->Value.cbData && j % 8)
                 {
@@ -4229,7 +4237,7 @@
         GetWindowRect(GetDlgItem(hwnd, IDCANCEL), &rc);
         MapWindowPoints( 0, hwnd, (POINT *)&rc, 2 );
         /* hide the cancel button.. */
-        ShowWindow(GetDlgItem(hwnd, IDCANCEL), SW_HIDE);
+        ShowWindow(GetDlgItem(hwnd, IDCANCEL), FALSE);
         /* and move the OK button to the cancel button's original position. */
         SetWindowPos(GetDlgItem(hwnd, IDOK), 0, rc.left, rc.top, 0, 0,
                      SWP_NOSIZE | SWP_NOZORDER | SWP_NOACTIVATE | SWP_NOREDRAW );
@@ -4241,10 +4249,11 @@
 static BOOL show_cert_dialog(PCCRYPTUI_VIEWCERTIFICATE_STRUCTW pCertViewInfo,
  CRYPT_PROVIDER_CERT *provCert, BOOL *pfPropertiesChanged)
 {
+    static const WCHAR riched[] = { 'r','i','c','h','e','d','2','0',0 };
     DWORD nPages;
     PROPSHEETPAGEW *pages;
     BOOL ret = FALSE;
-    HMODULE lib = LoadLibraryW(L"riched20");
+    HMODULE lib = LoadLibraryW(riched);
 
     nPages = pCertViewInfo->cPropSheetPages + 1; /* one for the General tab */
     if (!(pCertViewInfo->dwFlags & CRYPTUI_HIDE_DETAILPAGE))
@@ -4489,7 +4498,12 @@
 
 static HCERTSTORE choose_store_for_cert(PCCERT_CONTEXT cert)
 {
-    const WCHAR *storeName = is_ca_cert(cert, TRUE) ? L"CA" : L"AddressBook";
+    LPCWSTR storeName;
+
+    if (is_ca_cert(cert, TRUE))
+        storeName = ca;
+    else
+        storeName = addressBook;
     return CertOpenStore(CERT_STORE_PROV_SYSTEM_W, 0, 0,
      CERT_SYSTEM_STORE_CURRENT_USER, storeName);
 }
@@ -4533,7 +4547,7 @@
     else
     {
         if (!(store = CertOpenStore(CERT_STORE_PROV_SYSTEM_W, 0, 0,
-         CERT_SYSTEM_STORE_CURRENT_USER, L"CA")))
+         CERT_SYSTEM_STORE_CURRENT_USER, ca)))
         {
             WARN("unable to open certificate store\n");
             return FALSE;
@@ -4558,8 +4572,10 @@
     if (hDestCertStore) store = hDestCertStore;
     else
     {
+        static const WCHAR trust[] = { 'T','r','u','s','t',0 };
+
         if (!(store = CertOpenStore(CERT_STORE_PROV_SYSTEM_W, 0, 0,
-         CERT_SYSTEM_STORE_CURRENT_USER, L"Trust")))
+         CERT_SYSTEM_STORE_CURRENT_USER, trust)))
         {
             WARN("unable to open certificate store\n");
             return FALSE;
@@ -4832,19 +4848,30 @@
     return ret;
 }
 
+static const WCHAR filter_cert[] = { '*','.','c','e','r',';','*','.',
+ 'c','r','t',0 };
+static const WCHAR filter_pfx[] = { '*','.','p','f','x',';','*','.',
+ 'p','1','2',0 };
+static const WCHAR filter_crl[] = { '*','.','c','r','l',0 };
+static const WCHAR filter_ctl[] = { '*','.','s','t','l',0 };
+static const WCHAR filter_serialized_store[] = { '*','.','s','s','t',0 };
+static const WCHAR filter_cms[] = { '*','.','s','p','c',';','*','.',
+ 'p','7','b',0 };
+static const WCHAR filter_all[] = { '*','.','*',0 };
+
 static const struct
 {
     int     id;
     DWORD   allowFlags;
     LPCWSTR filter;
 } import_filters[] = {
- { IDS_IMPORT_FILTER_CERT, CRYPTUI_WIZ_IMPORT_ALLOW_CERT, L"*.cer;*.crt" },
- { IDS_IMPORT_FILTER_PFX, 0, L"*.pfx;*.p12" },
- { IDS_IMPORT_FILTER_CRL, CRYPTUI_WIZ_IMPORT_ALLOW_CRL, L"*.crl" },
- { IDS_IMPORT_FILTER_CTL, CRYPTUI_WIZ_IMPORT_ALLOW_CTL, L"*.stl" },
- { IDS_IMPORT_FILTER_SERIALIZED_STORE, 0, L"*.sst" },
- { IDS_IMPORT_FILTER_CMS, 0, L"*.spc;*.p7b" },
- { IDS_IMPORT_FILTER_ALL, 0, L"*.*" },
+ { IDS_IMPORT_FILTER_CERT, CRYPTUI_WIZ_IMPORT_ALLOW_CERT, filter_cert },
+ { IDS_IMPORT_FILTER_PFX, 0, filter_pfx },
+ { IDS_IMPORT_FILTER_CRL, CRYPTUI_WIZ_IMPORT_ALLOW_CRL, filter_crl },
+ { IDS_IMPORT_FILTER_CTL, CRYPTUI_WIZ_IMPORT_ALLOW_CTL, filter_ctl },
+ { IDS_IMPORT_FILTER_SERIALIZED_STORE, 0, filter_serialized_store },
+ { IDS_IMPORT_FILTER_CMS, 0, filter_cms },
+ { IDS_IMPORT_FILTER_ALL, 0, filter_all },
 };
 
 static WCHAR *make_import_file_filter(DWORD dwFlags)
@@ -5072,21 +5099,23 @@
         {
             SendMessageW(GetDlgItem(hwnd, IDC_IMPORT_AUTO_STORE), BM_CLICK,
              0, 0);
-            EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_STORE), TRUE);
-            EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_BROWSE_STORE), TRUE);
-            EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_SPECIFY_STORE),
-             !(data->dwFlags & CRYPTUI_WIZ_IMPORT_NO_CHANGE_DEST_STORE));
+            EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_STORE), FALSE);
+            EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_BROWSE_STORE), FALSE);
+            EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_SPECIFY_STORE), FALSE);
         }
         else
         {
+            WCHAR storeTitle[MAX_STRING_LEN];
+
             SendMessageW(GetDlgItem(hwnd, IDC_IMPORT_SPECIFY_STORE), BM_CLICK,
              0, 0);
             EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_STORE), TRUE);
             EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_BROWSE_STORE), TRUE);
             EnableWindow(GetDlgItem(hwnd, IDC_IMPORT_SPECIFY_STORE),
              !(data->dwFlags & CRYPTUI_WIZ_IMPORT_NO_CHANGE_DEST_STORE));
-            LoadStringW(hInstance, IDS_IMPORT_DEST_DETERMINED, selected_store_title, ARRAY_SIZE(selected_store_title));
-            SendMessageW(GetDlgItem(hwnd, IDC_IMPORT_STORE), WM_SETTEXT, 0, (LPARAM)selected_store_title);
+            LoadStringW(hInstance, IDS_IMPORT_DEST_DETERMINED, storeTitle, ARRAY_SIZE(storeTitle));
+            SendMessageW(GetDlgItem(hwnd, IDC_IMPORT_STORE), WM_SETTEXT,
+             0, (LPARAM)storeTitle);
         }
         break;
     }
@@ -5150,7 +5179,12 @@
             selectInfo.pfnSelectedStoreCallback = NULL;
             if ((store = CryptUIDlgSelectStoreW(&selectInfo)))
             {
-                SendMessageW(GetDlgItem(hwnd, IDC_IMPORT_STORE), WM_SETTEXT, 0, (LPARAM)selected_store_title);
+                WCHAR storeTitle[MAX_STRING_LEN];
+
+                LoadStringW(hInstance, IDS_IMPORT_DEST_DETERMINED, storeTitle,
+                 ARRAY_SIZE(storeTitle));
+                SendMessageW(GetDlgItem(hwnd, IDC_IMPORT_STORE), WM_SETTEXT,
+                 0, (LPARAM)storeTitle);
                 data->hDestCertStore = store;
                 data->freeDest = TRUE;
             }
@@ -5178,12 +5212,11 @@
     if (data->autoDest)
         LoadStringW(hInstance, IDS_IMPORT_DEST_AUTOMATIC, text, ARRAY_SIZE(text));
     else
-        item.pszText = selected_store_title;
+        LoadStringW(hInstance, IDS_IMPORT_DEST_DETERMINED, text, ARRAY_SIZE(text));
     SendMessageW(lv, LVM_SETITEMTEXTW, item.iItem, (LPARAM)&item);
     item.iItem = SendMessageW(lv, LVM_GETITEMCOUNT, 0, 0);
     item.iSubItem = 0;
     LoadStringW(hInstance, IDS_IMPORT_CONTENT, text, ARRAY_SIZE(text));
-    item.pszText = text;
     SendMessageW(lv, LVM_INSERTITEMW, 0, (LPARAM)&item);
     switch (data->contentType)
     {
@@ -5352,16 +5385,15 @@
 
     data.dwFlags = dwFlags;
     data.pwszWizardTitle = pwszWizardTitle;
-    data.fileName = NULL;
     if (pImportSrc)
     {
         memcpy(&data.importSrc, pImportSrc, sizeof(data.importSrc));
-        if (pImportSrc->dwSubjectChoice == CRYPTUI_WIZ_IMPORT_SUBJECT_FILE)
-            data.fileName = (LPWSTR)pImportSrc->pwszFileName;
+        data.fileName = (LPWSTR)pImportSrc->pwszFileName;
     }
     else
     {
         memset(&data.importSrc, 0, sizeof(data.importSrc));
+        data.fileName = NULL;
     }
     data.freeSource = FALSE;
     data.hDestCertStore = hDestCertStore;
@@ -5941,6 +5973,12 @@
 static LPWSTR export_append_extension(const struct ExportWizData *data,
  LPWSTR fileName)
 {
+    static const WCHAR cer[] = { '.','c','e','r',0 };
+    static const WCHAR crl[] = { '.','c','r','l',0 };
+    static const WCHAR ctl[] = { '.','c','t','l',0 };
+    static const WCHAR p7b[] = { '.','p','7','b',0 };
+    static const WCHAR pfx[] = { '.','p','f','x',0 };
+    static const WCHAR sst[] = { '.','s','s','t',0 };
     LPCWSTR extension;
     LPWSTR dot;
     BOOL appendExtension;
@@ -5948,25 +5986,25 @@
     switch (data->contextInfo.dwExportFormat)
     {
     case CRYPTUI_WIZ_EXPORT_FORMAT_PKCS7:
-        extension = L".p7b";
+        extension = p7b;
         break;
     case CRYPTUI_WIZ_EXPORT_FORMAT_PFX:
-        extension = L".pfx";
+        extension = pfx;
         break;
     default:
         switch (data->exportInfo.dwSubjectChoice)
         {
         case CRYPTUI_WIZ_EXPORT_CRL_CONTEXT:
-            extension = L".crl";
+            extension = crl;
             break;
         case CRYPTUI_WIZ_EXPORT_CTL_CONTEXT:
-            extension = L".ctl";
+            extension = ctl;
             break;
         case CRYPTUI_WIZ_EXPORT_CERT_STORE:
-            extension = L".sst";
+            extension = sst;
             break;
         default:
-            extension = L".cer";
+            extension = cer;
         }
     }
     dot = wcsrchr(fileName, '.');
@@ -6059,6 +6097,13 @@
     return ret;
 }
 
+static const WCHAR export_filter_cert[] = { '*','.','c','e','r',0 };
+static const WCHAR export_filter_crl[] = { '*','.','c','r','l',0 };
+static const WCHAR export_filter_ctl[] = { '*','.','s','t','l',0 };
+static const WCHAR export_filter_cms[] = { '*','.','p','7','b',0 };
+static const WCHAR export_filter_pfx[] = { '*','.','p','f','x',0 };
+static const WCHAR export_filter_sst[] = { '*','.','s','s','t',0 };
+
 static WCHAR *make_export_file_filter(DWORD exportFormat, DWORD subjectChoice)
 {
     int baseLen, allLen, totalLen = 2, baseID;
@@ -6069,41 +6114,41 @@
     {
     case CRYPTUI_WIZ_EXPORT_FORMAT_BASE64:
         baseID = IDS_EXPORT_FILTER_BASE64_CERT;
-        filterStr = L"*.cer";
+        filterStr = export_filter_cert;
         break;
     case CRYPTUI_WIZ_EXPORT_FORMAT_PFX:
         baseID = IDS_EXPORT_FILTER_PFX;
-        filterStr = L"*.pfx";
+        filterStr = export_filter_pfx;
         break;
     case CRYPTUI_WIZ_EXPORT_FORMAT_PKCS7:
         baseID = IDS_EXPORT_FILTER_CMS;
-        filterStr = L"*.p7b";
+        filterStr = export_filter_cms;
         break;
     default:
         switch (subjectChoice)
         {
         case CRYPTUI_WIZ_EXPORT_CRL_CONTEXT:
             baseID = IDS_EXPORT_FILTER_CRL;
-            filterStr = L"*.crl";
+            filterStr = export_filter_crl;
             break;
         case CRYPTUI_WIZ_EXPORT_CTL_CONTEXT:
             baseID = IDS_EXPORT_FILTER_CTL;
-            filterStr = L"*.stl";
+            filterStr = export_filter_ctl;
             break;
         case CRYPTUI_WIZ_EXPORT_CERT_STORE:
             baseID = IDS_EXPORT_FILTER_SERIALIZED_CERT_STORE;
-            filterStr = L"*.sst";
+            filterStr = export_filter_sst;
             break;
         default:
             baseID = IDS_EXPORT_FILTER_CERT;
-            filterStr = L"*.cer";
+            filterStr = export_filter_cert;
             break;
         }
     }
     baseLen = LoadStringW(hInstance, baseID, (LPWSTR)&baseFilter, 0);
     totalLen += baseLen + lstrlenW(filterStr) + 2;
     allLen = LoadStringW(hInstance, IDS_IMPORT_FILTER_ALL, (LPWSTR)&all, 0);
-    totalLen += allLen + ARRAY_SIZE(L"*.*") + 1;
+    totalLen += allLen + lstrlenW(filter_all) + 2;
     filter = malloc(totalLen * sizeof(WCHAR));
     if (filter)
     {
@@ -6118,8 +6163,8 @@
         memcpy(ptr, all, allLen * sizeof(WCHAR));
         ptr += allLen;
         *ptr++ = 0;
-        lstrcpyW(ptr, L"*.*");
-        ptr += ARRAY_SIZE(L"*.*");
+        lstrcpyW(ptr, filter_all);
+        ptr += lstrlenW(filter_all) + 1;
         *ptr++ = 0;
     }
     return filter;
