diff --git a/dlls/hhctrl.ocx/help.c b/dlls/hhctrl.ocx/help.c
--- a/dlls/hhctrl.ocx/help.c
+++ b/dlls/hhctrl.ocx/help.c
@@ -25,7 +25,6 @@
 #include "wingdi.h"
 #include "commctrl.h"
 #include "wininet.h"
-#include "exdispid.h"
 
 #include "wine/debug.h"
 
@@ -48,8 +47,6 @@
 #define TAB_RIGHT_PADDING   4
 #define TAB_MARGIN  8
 #define EDIT_HEIGHT         20
-#define BUTTON_HEIGHT       25
-#define BUTTON_WIDTH        65
 
 struct list window_list = LIST_INIT(window_list);
 
@@ -236,6 +233,10 @@
 
 static BOOL AppendFullPathURL(LPCWSTR file, LPWSTR buf, LPCWSTR index)
 {
+    static const WCHAR url_format[] =
+        {'m','k',':','@','M','S','I','T','S','t','o','r','e',':','%','s',':',':','%','s','%','s',0};
+    static const WCHAR slash[] = {'/',0};
+    static const WCHAR empty[] = {0};
     WCHAR full_path[MAX_PATH];
 
     TRACE("%s %p %s\n", debugstr_w(file), buf, debugstr_w(index));
@@ -245,7 +246,7 @@
         return FALSE;
     }
 
-    wsprintfW(buf, L"mk:@MSITStore:%s::%s%s", full_path, (!index || index[0] == '/') ? L"" : L"/", index);
+    wsprintfW(buf, url_format, full_path, (!index || index[0] == '/') ? empty : slash, index);
     return TRUE;
 }
 
@@ -261,21 +262,12 @@
     return SUCCEEDED(navigate_url(info, buf));
 }
 
-static BOOL is_chm(WCHAR *url)
-{
-    const WCHAR *prefix = L"mk:@MSITStore:";
-    return !wcsncmp(url, prefix, wcslen(prefix));
-}
-
-static void DoSyncContent(HHInfo *info)
-{
-    const WCHAR *index;
+static void DoSync(HHInfo *info)
+{
+    WCHAR buf[INTERNET_MAX_URL_LENGTH];
     HRESULT hres;
     BSTR url;
 
-    if (info->current_tab != TAB_CONTENTS)
-        return;
-
     hres = IWebBrowser2_get_LocationURL(info->web_browser->web_browser, &url);
 
     if (FAILED(hres))
@@ -284,131 +276,34 @@
         return;
     }
 
-    /* If we're not currently viewing a page in a .chm file, abort */
-    if (!is_chm(url))
+    /* If we're not currently viewing a page in the active .chm file, abort */
+    if ((!AppendFullPathURL(info->WinType.pszFile, buf, NULL)) || (lstrlenW(buf) > lstrlenW(url)))
     {
         SysFreeString(url);
         return;
     }
 
-    index = wcsstr(url, L"::/");
-
-    if (index)
-        ActivateContentTopic(info->tabs[TAB_CONTENTS].hwnd, index + 3, info->content); /* skip over ::/ */
+    if (lstrcmpiW(buf, url) > 0)
+    {
+        static const WCHAR delimW[] = {':',':','/',0};
+        const WCHAR *index;
+
+        index = wcsstr(url, delimW);
+
+        if (index)
+            ActivateContentTopic(info->tabs[TAB_CONTENTS].hwnd, index + 3, info->content); /* skip over ::/ */
+    }
 
     SysFreeString(url);
 }
 
-static HRESULT WINAPI WebBrowserEvents2_QueryInterface(IDispatch *iface, REFIID riid, void **v)
-{
-    *v = NULL;
-
-    if (IsEqualGUID(&IID_IDispatch, riid) || IsEqualGUID(&IID_IUnknown, riid))
-    {
-        *v = iface;
-        IDispatch_AddRef(iface);
-        return S_OK;
-    }
-
-    return E_NOINTERFACE;
-}
-
-static inline WebBrowserEvents2Impl *impl_from_IDispatch(IDispatch *iface)
-{
-    return CONTAINING_RECORD(iface, WebBrowserEvents2Impl, WebBrowserEvents2Impl_iface);
-}
-
-static ULONG WINAPI WebBrowserEvents2_AddRef(IDispatch *iface)
-{
-    WebBrowserEvents2Impl *impl = impl_from_IDispatch(iface);
-    return InterlockedIncrement(&impl->ref);
-}
-
-static ULONG WINAPI WebBrowserEvents2_Release(IDispatch *iface)
-{
-    WebBrowserEvents2Impl *impl = impl_from_IDispatch(iface);
-    ULONG ref = InterlockedDecrement(&impl->ref);
-
-    if (!ref)
-        free(impl);
-    return ref;
-}
-
-static HRESULT WINAPI WebBrowserEvents2_GetTypeInfoCount(IDispatch *iface, UINT *pctinfo)
-{
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI WebBrowserEvents2_GetTypeInfo(IDispatch *iface, UINT iTInfo, LCID lcid,
-        ITypeInfo **ppTInfo)
-{
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI WebBrowserEvents2_GetIDsOfNames(IDispatch *iface, REFIID riid, LPOLESTR *rgszNames,
-        UINT cNames, LCID lcid, DISPID *rgDispId)
-{
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI WebBrowserEvents2_Invoke(IDispatch *iface, DISPID dispIdMember, REFIID riid,
-        LCID lcid, WORD wFlags, DISPPARAMS *pDispParams, VARIANT *pVarResult,
-        EXCEPINFO *pExcepInfo, UINT *puArgErr)
-{
-    if (dispIdMember == DISPID_NAVIGATECOMPLETE2)
-    {
-        WebBrowserEvents2Impl *impl = impl_from_IDispatch(iface);
-        DoSyncContent((HHInfo *)impl->info);
-    }
-    return S_OK;
-}
-
-static const IDispatchVtbl WebBrowserEvents2Vtbl =
-{
-    WebBrowserEvents2_QueryInterface,
-    WebBrowserEvents2_AddRef,
-    WebBrowserEvents2_Release,
-    WebBrowserEvents2_GetTypeInfoCount,
-    WebBrowserEvents2_GetTypeInfo,
-    WebBrowserEvents2_GetIDsOfNames,
-    WebBrowserEvents2_Invoke
+/* Size Bar */
+
+#define SIZEBAR_WIDTH   4
+
+static const WCHAR szSizeBarClass[] = {
+    'H','H',' ','S','i','z','e','B','a','r',0
 };
-
-static void hook_WebBrowserEvents2(HHInfo *info, BOOL init)
-{
-    IConnectionPointContainer *container;
-    IConnectionPoint *point;
-    HRESULT hres;
-
-    if (!info->web_browser || !info->web_browser->web_browser)
-        return;
-
-    hres = IWebBrowser2_QueryInterface(info->web_browser->web_browser, &IID_IConnectionPointContainer, (void **)&container);
-    if (FAILED(hres))
-        return;
-
-    hres = IConnectionPointContainer_FindConnectionPoint(container, &DIID_DWebBrowserEvents2, &point);
-    IConnectionPointContainer_Release(container);
-    if (FAILED(hres))
-        return;
-
-    if (init)
-    {
-        info->web_browser->WebBrowser_events_sink = malloc(sizeof(*(info->web_browser->WebBrowser_events_sink)));
-        info->web_browser->WebBrowser_events_sink->WebBrowserEvents2Impl_iface.lpVtbl = &WebBrowserEvents2Vtbl;
-        info->web_browser->WebBrowser_events_sink->info = (struct HHInfo *)info;
-        info->web_browser->WebBrowser_events_sink->ref = 1;
-        IConnectionPoint_Advise(point, (IUnknown *)info->web_browser->WebBrowser_events_sink, &info->web_browser->WebBrowser_events_sink->cookie);
-    }
-    else
-        IConnectionPoint_Unadvise(point, info->web_browser->WebBrowser_events_sink->cookie);
-
-    IConnectionPoint_Release(point);
-}
-
-/* Size Bar */
-
-#define SIZEBAR_WIDTH   4
 
 /* Draw the SizeBar */
 static void SB_OnPaint(HWND hWnd)
@@ -503,7 +398,7 @@
     wcex.hCursor        = LoadCursorW(NULL, (LPCWSTR)IDC_SIZEWE);
     wcex.hbrBackground  = (HBRUSH)(COLOR_MENU + 1);
     wcex.lpszMenuName   = NULL;
-    wcex.lpszClassName  = L"HH SizeBar";
+    wcex.lpszClassName  = szSizeBarClass;
     wcex.hIconSm        = LoadIconW(NULL, (LPCWSTR)IDI_APPLICATION);
 
     RegisterClassExW(&wcex);
@@ -533,7 +428,7 @@
 
     SB_GetSizeBarRect(pHHInfo, &rc);
 
-    hWnd = CreateWindowExW(dwExStyles, L"HH SizeBar", L"", dwStyles,
+    hWnd = CreateWindowExW(dwExStyles, szSizeBarClass, szEmpty, dwStyles,
                            rc.left, rc.top, rc.right, rc.bottom,
                            hwndParent, NULL, hhctrl_hinstance, NULL);
     if (!hWnd)
@@ -547,6 +442,10 @@
 }
 
 /* Child Window */
+
+static const WCHAR szChildClass[] = {
+    'H','H',' ','C','h','i','l','d',0
+};
 
 static LRESULT Child_OnPaint(HWND hWnd)
 {
@@ -620,19 +519,11 @@
         int scroll_width = GetSystemMetrics(SM_CXVSCROLL);
         int border_width = GetSystemMetrics(SM_CXBORDER);
         int edge_width = GetSystemMetrics(SM_CXEDGE);
-
-        int right_pos = rect.right - TAB_MARGIN - BUTTON_WIDTH;
         int top_pos = 0;
 
         SetWindowPos(info->search.hwndEdit, NULL, 0, top_pos, width,
                       EDIT_HEIGHT, SWP_NOZORDER | SWP_NOACTIVATE);
         top_pos += EDIT_HEIGHT + TAB_MARGIN;
-
-        if (0 > right_pos)
-            right_pos = 0;
-        SetWindowPos(info->search.hwndSearchBtn, NULL, right_pos, top_pos, BUTTON_WIDTH, BUTTON_HEIGHT, SWP_NOZORDER | SWP_NOACTIVATE);
-        top_pos += BUTTON_HEIGHT + TAB_MARGIN;
-
         SetWindowPos(info->search.hwndList, NULL, 0, top_pos, width,
                       height-top_pos, SWP_NOZORDER | SWP_NOACTIVATE);
         /* Resize the tab widget column to perfectly fit the tab window and
@@ -699,34 +590,7 @@
     if(info->tabs[info->current_tab].hwnd)
         ShowWindow(info->tabs[info->current_tab].hwnd, SW_SHOW);
 
-    if (info->current_tab == TAB_CONTENTS)
-        DoSyncContent(info);
-
     return 0;
-}
-
-static BOOL is_current_page(HHInfo *info, const WCHAR *local)
-{
-    WCHAR *url;
-    WCHAR *current_page;
-    HRESULT res;
-
-    res = IWebBrowser2_get_LocationURL(info->web_browser->web_browser, &url);
-    if (FAILED(res))
-        return FALSE;
-
-    if (is_chm(url))
-    {
-        current_page = wcsstr(url, L"::/");
-        if (current_page && !lstrcmpW(local, current_page + 3))
-            return TRUE;
-    }else
-    {
-        if (!lstrcmpW(local, url))
-            return TRUE;
-    }
-
-    return FALSE;
 }
 
 static LRESULT OnTopicChange(HHInfo *info, void *user_data)
@@ -745,9 +609,6 @@
         citer = (ContentItem *) user_data;
         name = citer->name;
         local = citer->local;
-        if (is_current_page(info, local))
-            return 0;
-
         while(citer) {
             if(citer->merge.chm_file) {
                 chmfile = citer->merge.chm_file;
@@ -807,7 +668,7 @@
         return 0;
     }
 
-    TRACE("name %s local %s\n", debugstr_w(name), debugstr_w(local));
+    TRACE("name %s loal %s\n", debugstr_w(name), debugstr_w(local));
 
     NavigateToChm(info, chmfile, local);
     return 0;
@@ -827,25 +688,6 @@
         SendMessageW(GetParent(GetParent(hWnd)), WM_NOTIFY, wParam, (LPARAM)&nmhdr);
     }
     return editWndProc(hWnd, message, wParam, lParam);
-}
-
-static void do_search(HHInfo *info)
-{
-    char needle[100];
-    DWORD i, len;
-
-    len = GetWindowTextA(info->search.hwndEdit, needle, sizeof(needle));
-    if(!len)
-    {
-        FIXME("Unable to get search text.\n");
-        return;
-    }
-    /* Convert the requested text for comparison later against the
-     * lower case version of HTML file contents.
-     */
-    for(i=0;i<len;i++)
-        needle[i] = tolower(needle[i]);
-    InitSearch(info, needle);
 }
 
 static LRESULT CALLBACK Child_WndProc(HWND hWnd, UINT message, WPARAM wParam, LPARAM lParam)
@@ -910,7 +752,21 @@
             }
             case TAB_SEARCH: {
                 if(nmhdr->hwndFrom == info->search.hwndEdit) {
-                    do_search(info);
+                    char needle[100];
+                    DWORD i, len;
+
+                    len = GetWindowTextA(info->search.hwndEdit, needle, sizeof(needle));
+                    if(!len)
+                    {
+                        FIXME("Unable to get search text.\n");
+                        return 0;
+                    }
+                    /* Convert the requested text for comparison later against the
+                     * lower case version of HTML file contents.
+                     */
+                    for(i=0;i<len;i++)
+                        needle[i] = tolower(needle[i]);
+                    InitSearch(info, needle);
                     return 0;
                 }else if(nmhdr->hwndFrom == info->search.hwndList) {
                     HWND hwndList = info->search.hwndList;
@@ -929,12 +785,6 @@
         }
         break;
     }
-    case WM_COMMAND:
-    {
-        HHInfo *info = (HHInfo*)GetWindowLongPtrW(hWnd, 0);
-        if (info->current_tab == TAB_SEARCH && LOWORD(wParam) == IDC_SEARCH_BTN && HIWORD(wParam) == BN_CLICKED)
-            do_search(info);
-    }
     default:
         return DefWindowProcW(hWnd, message, wParam, lParam);
     }
@@ -956,7 +806,7 @@
     wcex.hCursor        = LoadCursorW(NULL, (LPCWSTR)IDC_ARROW);
     wcex.hbrBackground  = (HBRUSH)(COLOR_BTNFACE + 1);
     wcex.lpszMenuName   = NULL;
-    wcex.lpszClassName  = L"HH Child";
+    wcex.lpszClassName  = szChildClass;
     wcex.hIconSm        = LoadIconW(NULL, (LPCWSTR)IDI_APPLICATION);
 
     RegisterClassExW(&wcex);
@@ -1043,7 +893,7 @@
             ExpandContract(info);
             break;
         case IDTB_SYNC:
-            DoSyncContent(info);
+            DoSync(info);
             break;
         case IDTB_OPTIONS:
             DisplayPopupMenu(info);
@@ -1242,7 +1092,7 @@
 
     NP_GetNavigationRect(info, &rc);
 
-    hWnd = CreateWindowExW(dwExStyles, L"HH Child", szEmpty, dwStyles,
+    hWnd = CreateWindowExW(dwExStyles, szChildClass, szEmpty, dwStyles,
                            rc.left, rc.top, rc.right, rc.bottom,
                            hwndParent, NULL, hhctrl_hinstance, NULL);
     if (!hWnd)
@@ -1312,7 +1162,7 @@
 
     HP_GetHTMLRect(pHHInfo, &rc);
 
-    hWnd = CreateWindowExW(dwExStyles, L"HH Child", szEmpty, dwStyles,
+    hWnd = CreateWindowExW(dwExStyles, szChildClass, szEmpty, dwStyles,
                            rc.left, rc.top, rc.right, rc.bottom,
                            hwndParent, NULL, hhctrl_hinstance, NULL);
     if (!hWnd)
@@ -1320,8 +1170,6 @@
 
     if (!InitWebBrowser(pHHInfo, hWnd))
         return FALSE;
-
-    hook_WebBrowserEvents2(pHHInfo, TRUE);
 
     /* store the pointer to the HH info struct */
     SetWindowLongPtrW(hWnd, 0, (LONG_PTR)pHHInfo);
@@ -1393,14 +1241,14 @@
 
 static BOOL AddSearchTab(HHInfo *info)
 {
-    HWND hwndList, hwndEdit, hwndContainer, hwndSearchBtn;
+    HWND hwndList, hwndEdit, hwndContainer;
     char hidden_column[] = "Column";
     WNDPROC editWndProc;
     LVCOLUMNA lvc;
 
     if(info->tabs[TAB_SEARCH].id == -1)
         return TRUE; /* No "Search" tab */
-    hwndContainer = CreateWindowExW(WS_EX_CONTROLPARENT, L"HH Child", szEmpty,
+    hwndContainer = CreateWindowExW(WS_EX_CONTROLPARENT, szChildClass, szEmpty,
                                     WS_CHILD, 0, 0, 0, 0, info->WinType.hwndNavigation,
                                     NULL, hhctrl_hinstance, NULL);
     if(!hwndContainer) {
@@ -1425,15 +1273,6 @@
         return FALSE;
     }
     SetWindowLongPtrW(hwndEdit, GWLP_USERDATA, (LONG_PTR)editWndProc);
-    hwndSearchBtn = CreateWindowExW(WS_EX_NOPARENTNOTIFY, WC_BUTTONW, szEmpty,
-                                    WS_CHILD | WS_VISIBLE | WS_TABSTOP | WS_GROUP,
-                                    0, 0, 0, 0, hwndContainer, IDC_SEARCH_BTN, hhctrl_hinstance, NULL);
-    if (SendMessageW(hwndSearchBtn, WM_SETFONT, (WPARAM) info->hFont, (LPARAM) FALSE) == -1)
-    {
-        ERR("Could not set font for \"List Topics\" button.\n");
-        return FALSE;
-    }
-    SetWindowTextW(hwndSearchBtn, L"&List Topics");
     hwndList = CreateWindowExW(WS_EX_CLIENTEDGE, WC_LISTVIEWW, szEmpty,
                                WS_CHILD | WS_VISIBLE | WS_BORDER | LVS_SINGLESEL
                                 | LVS_REPORT | LVS_NOCOLUMNHEADER, 0, 0, 0, 0,
@@ -1454,7 +1293,6 @@
     info->search.hwndEdit = hwndEdit;
     info->search.hwndList = hwndList;
     info->search.hwndContainer = hwndContainer;
-    info->search.hwndSearchBtn = hwndSearchBtn;
     info->tabs[TAB_SEARCH].hwnd = hwndContainer;
 
     SetWindowLongPtrW(hwndContainer, 0, (LONG_PTR)info);
@@ -1567,9 +1405,11 @@
 
 static BOOL AddIndexPopup(HHInfo *info)
 {
+    static const WCHAR szPopupChildClass[] = {'H','H',' ','P','o','p','u','p',' ','C','h','i','l','d',0};
+    static const WCHAR windowCaptionW[] = {'S','e','l','e','c','t',' ','T','o','p','i','c',':',0};
+    static const WCHAR windowClassW[] = {'H','H',' ','P','o','p','u','p',0};
     HWND hwndList, hwndPopup, hwndCallback;
     char hidden_column[] = "Column";
-    WCHAR *window_title;
     WNDCLASSEXW wcex;
     LVCOLUMNA lvc;
 
@@ -1586,7 +1426,7 @@
     wcex.hCursor        = LoadCursorW(NULL, (LPCWSTR)IDC_ARROW);
     wcex.hbrBackground  = (HBRUSH)(COLOR_MENU + 1);
     wcex.lpszMenuName   = NULL;
-    wcex.lpszClassName  = L"HH Popup";
+    wcex.lpszClassName  = windowClassW;
     wcex.hIconSm        = LoadIconW(NULL, (LPCWSTR)IDI_APPLICATION);
     RegisterClassExW(&wcex);
 
@@ -1600,24 +1440,22 @@
     wcex.hCursor        = LoadCursorW(NULL, (LPCWSTR)IDC_ARROW);
     wcex.hbrBackground  = (HBRUSH)(COLOR_BTNFACE + 1);
     wcex.lpszMenuName   = NULL;
-    wcex.lpszClassName  = L"HH Popup Child";
+    wcex.lpszClassName  = szPopupChildClass;
     wcex.hIconSm        = LoadIconW(NULL, (LPCWSTR)IDI_APPLICATION);
     RegisterClassExW(&wcex);
 
-    window_title = HH_LoadString(IDS_SELECT_TOPIC);
     hwndPopup = CreateWindowExW(WS_EX_LEFT | WS_EX_LTRREADING | WS_EX_APPWINDOW
                                  | WS_EX_WINDOWEDGE | WS_EX_RIGHTSCROLLBAR,
-                                L"HH Popup", window_title, WS_POPUPWINDOW
+                                windowClassW, windowCaptionW, WS_POPUPWINDOW
                                  | WS_OVERLAPPEDWINDOW | WS_VISIBLE
                                  | WS_CLIPSIBLINGS | WS_CLIPCHILDREN, CW_USEDEFAULT,
                                 CW_USEDEFAULT, 300, 200, info->WinType.hwndHelp,
                                 NULL, hhctrl_hinstance, NULL);
-    free(window_title);
     if (!hwndPopup)
         return FALSE;
 
     hwndCallback = CreateWindowExW(WS_EX_LEFT | WS_EX_LTRREADING | WS_EX_RIGHTSCROLLBAR,
-                                   L"HH Popup Child", szEmpty, WS_CHILDWINDOW | WS_VISIBLE,
+                                   szPopupChildClass, szEmpty, WS_CHILDWINDOW | WS_VISIBLE,
                                    0, 0, 0, 0,
                                    hwndPopup, NULL, hhctrl_hinstance, NULL);
     if (!hwndCallback)
@@ -1772,6 +1610,10 @@
     DWORD x, y, width = 0, height = 0;
     LPCWSTR caption;
 
+    static const WCHAR windowClassW[] = {
+        'H','H',' ', 'P','a','r','e','n','t',0
+    };
+
     wcex.cbSize         = sizeof(WNDCLASSEXW);
     wcex.style          = CS_HREDRAW | CS_VREDRAW;
     wcex.lpfnWndProc    = Help_WndProc;
@@ -1782,7 +1624,7 @@
     wcex.hCursor        = LoadCursorW(NULL, (LPCWSTR)IDC_ARROW);
     wcex.hbrBackground  = (HBRUSH)(COLOR_MENU + 1);
     wcex.lpszMenuName   = NULL;
-    wcex.lpszClassName  = L"HH Parent";
+    wcex.lpszClassName  = windowClassW;
     wcex.hIconSm        = LoadIconW(NULL, (LPCWSTR)IDI_APPLICATION);
 
     RegisterClassExW(&wcex);
@@ -1835,7 +1677,7 @@
     caption = info->WinType.pszCaption;
     if (!*caption) caption = info->pCHMInfo->defTitle;
 
-    hWnd = CreateWindowExW(dwExStyles, L"HH Parent", caption, dwStyles, x, y, width, height,
+    hWnd = CreateWindowExW(dwExStyles, windowClassW, caption, dwStyles, x, y, width, height,
                            info->WinType.hwndCaller, NULL, hhctrl_hinstance, NULL);
     if (!hWnd)
         return FALSE;
@@ -1960,7 +1802,6 @@
     if (info->pCHMInfo)
         CloseCHM(info->pCHMInfo);
 
-    hook_WebBrowserEvents2(info, FALSE);
     ReleaseWebBrowser(info);
     ReleaseContent(info);
     ReleaseIndex(info);
