diff --git a/dlls/winevulkan/loader.c b/dlls/winevulkan/loader.c
--- a/dlls/winevulkan/loader.c
+++ b/dlls/winevulkan/loader.c
@@ -20,6 +20,10 @@
 #include <stdlib.h>
 #include "vulkan_loader.h"
 #include "winreg.h"
+#include "ntuser.h"
+#include "initguid.h"
+#include "devguid.h"
+#include "setupapi.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(vulkan);
 
@@ -28,9 +32,26 @@
  */
 #define WINE_VULKAN_ICD_VERSION 4
 
+DEFINE_DEVPROPKEY(DEVPROPKEY_GPU_LUID, 0x60b193cb, 0x5276, 0x4d0f, 0x96, 0xfc, 0xf1, 0x73, 0xab, 0xad, 0x3e, 0xc6, 2);
+DEFINE_DEVPROPKEY(WINE_DEVPROPKEY_GPU_VULKAN_UUID, 0x233a9ef3, 0xafc4, 0x4abd, 0xb5, 0x64, 0xc3, 0x2f, 0x21, 0xf1, 0x53, 0x5c, 2);
+
 static HINSTANCE hinstance;
-static BOOL get_device_proc_addr_instance_procs = FALSE;
-static struct vulkan_instance_extensions instance_extensions; /* supported client instance extensions */
+
+static void *wine_vk_get_global_proc_addr(const char *name);
+
+#define wine_vk_find_struct(s, t) wine_vk_find_struct_((void *)s, VK_STRUCTURE_TYPE_##t)
+static void *wine_vk_find_struct_(void *s, VkStructureType t)
+{
+    VkBaseOutStructure *header;
+
+    for (header = s; header; header = header->pNext)
+    {
+        if (header->sType == t)
+            return header;
+    }
+
+    return NULL;
+}
 
 VkResult WINAPI vkEnumerateInstanceLayerProperties(uint32_t *count, VkLayerProperties *properties)
 {
@@ -107,14 +128,6 @@
         return NULL;
     }
 
-    if (instance->extensions.has_VK_KHR_win32_surface)
-    {
-        if (!strcmp(name, "vkCreateWin32SurfaceKHR"))
-            return (PFN_vkVoidFunction)vkCreateWin32SurfaceKHR;
-        if (!strcmp(name, "vkGetPhysicalDeviceWin32PresentationSupportKHR"))
-            return (PFN_vkVoidFunction)vkGetPhysicalDeviceWin32PresentationSupportKHR;
-    }
-
     if (!is_available_instance_function(instance, name))
         return NULL;
 
@@ -140,30 +153,6 @@
     /* The spec leaves return value undefined for a NULL device, let's just return NULL. */
     if (!device || !name)
         return NULL;
-
-    if (device->extensions.has_VK_KHR_external_memory_win32)
-    {
-        if (!strcmp(name, "vkGetMemoryWin32HandleKHR"))
-            return (PFN_vkVoidFunction)vkGetMemoryWin32HandleKHR;
-        if (!strcmp(name, "vkGetMemoryWin32HandlePropertiesKHR"))
-            return (PFN_vkVoidFunction)vkGetMemoryWin32HandlePropertiesKHR;
-    }
-
-    if (device->extensions.has_VK_KHR_external_semaphore_win32)
-    {
-        if (!strcmp(name, "vkGetSemaphoreWin32HandleKHR"))
-            return (PFN_vkVoidFunction)vkGetSemaphoreWin32HandleKHR;
-        if (!strcmp(name, "vkImportSemaphoreWin32HandleKHR"))
-            return (PFN_vkVoidFunction)vkImportSemaphoreWin32HandleKHR;
-    }
-
-    if (device->extensions.has_VK_KHR_external_fence_win32)
-    {
-        if (!strcmp(name, "vkGetFenceWin32HandleKHR"))
-            return (PFN_vkVoidFunction)vkGetFenceWin32HandleKHR;
-        if (!strcmp(name, "vkImportFenceWin32HandleKHR"))
-            return (PFN_vkVoidFunction)vkImportFenceWin32HandleKHR;
-    }
 
     /* Per the spec, we are only supposed to return device functions as in functions
      * for which the first parameter is vkDevice or a child of vkDevice like a
@@ -188,7 +177,7 @@
      * https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/issues/2323
      * https://github.com/KhronosGroup/Vulkan-Docs/issues/655
      */
-    if (get_device_proc_addr_instance_procs
+    if ((device->quirks & WINEVULKAN_QUIRK_GET_DEVICE_PROC_ADDR)
         && ((func = wine_vk_get_instance_proc_addr(name))
              || (func = wine_vk_get_phys_dev_proc_addr(name))))
     {
@@ -338,14 +327,13 @@
 
 static BOOL WINAPI wine_vk_init(INIT_ONCE *once, void *param, void **context)
 {
-    struct init_params params =
-    {
-        .call_vulkan_debug_report_callback = (UINT_PTR)call_vulkan_debug_report_callback,
-        .call_vulkan_debug_utils_callback = (UINT_PTR)call_vulkan_debug_utils_callback,
-        .extensions = &instance_extensions,
+    struct vk_callback_funcs callback_funcs =
+    {
+        .call_vulkan_debug_report_callback = (ULONG_PTR)call_vulkan_debug_report_callback,
+        .call_vulkan_debug_utils_callback = (ULONG_PTR)call_vulkan_debug_utils_callback,
     };
 
-    return !__wine_init_unix_call() && !UNIX_CALL(init, &params);
+    return !__wine_init_unix_call() && !UNIX_CALL(init, &callback_funcs);
 }
 
 static BOOL  wine_vk_init_once(void)
@@ -353,34 +341,14 @@
     static INIT_ONCE init_once = INIT_ONCE_STATIC_INIT;
 
     return InitOnceExecuteOnce(&init_once, wine_vk_init, NULL, NULL);
-}
-
-static BOOL is_instance_extension_supported(const char *extension, struct vulkan_instance_extensions *extensions)
-{
-#define USE_VK_EXT(x) if (!strcmp(extension, #x)) return (extensions->has_ ## x = instance_extensions.has_ ## x);
-    ALL_VK_CLIENT_INSTANCE_EXTS
-#undef USE_VK_EXT
-    WARN("Extension %s is not supported.\n", debugstr_a(extension));
-    return FALSE;
-}
-
-static BOOL is_device_extension_supported(VkPhysicalDevice physical_device, const char *extension, struct vulkan_device_extensions *extensions)
-{
-#define USE_VK_EXT(x) if (!strcmp(extension, #x)) return (extensions->has_ ## x = physical_device->extensions.has_ ## x);
-    ALL_VK_CLIENT_DEVICE_EXTS
-#undef USE_VK_EXT
-    WARN("Extension %s is not supported.\n", debugstr_a(extension));
-    return FALSE;
 }
 
 VkResult WINAPI vkCreateInstance(const VkInstanceCreateInfo *create_info,
         const VkAllocationCallbacks *allocator, VkInstance *ret)
 {
-    struct vulkan_instance_extensions extensions = {0};
     struct vkCreateInstance_params params;
-    const VkApplicationInfo *app_info;
     struct VkInstance_T *instance;
-    uint32_t device_count = 8, i;
+    uint32_t phys_dev_count = 8, i;
     NTSTATUS status;
 
     TRACE("create_info %p, allocator %p, instance %p\n", create_info, allocator, ret);
@@ -388,52 +356,28 @@
     if (!wine_vk_init_once())
         return VK_ERROR_INITIALIZATION_FAILED;
 
-    if ((app_info = create_info->pApplicationInfo))
-    {
-        TRACE("Application name %s, application version %#x.\n", debugstr_a(app_info->pApplicationName), app_info->applicationVersion);
-        TRACE("Engine name %s, engine version %#x.\n", debugstr_a(app_info->pEngineName), app_info->engineVersion);
-        TRACE("API version %#x.\n", app_info->apiVersion);
-
-        if (app_info->pEngineName && !strcmp(app_info->pEngineName, "idTech"))
-            get_device_proc_addr_instance_procs = TRUE;
-    }
-
-    TRACE("Enabling %u client instance extensions\n", create_info->enabledExtensionCount);
-    for (uint32_t i = 0; i < create_info->enabledExtensionCount; i++)
-    {
-        const char *extension = create_info->ppEnabledExtensionNames[i];
-        if (!is_instance_extension_supported(extension, &extensions))
-            return VK_ERROR_EXTENSION_NOT_PRESENT;
-        TRACE("  - %s\n", extension);
-    }
-
     for (;;)
     {
-        if (!(instance = vulkan_client_object_create(FIELD_OFFSET(struct VkInstance_T, physical_device[device_count]))))
+        if (!(instance = vulkan_client_object_create(FIELD_OFFSET(struct VkInstance_T, phys_devs[phys_dev_count]))))
             return VK_ERROR_OUT_OF_HOST_MEMORY;
-        instance->physical_device_count = device_count;
-        for (i = 0; i < device_count; i++)
-            instance->physical_device[i].obj.loader_magic = VULKAN_ICD_MAGIC_VALUE;
-        instance->extensions = extensions;
-        *ret = instance;
+        instance->phys_dev_count = phys_dev_count;
+        for (i = 0; i < phys_dev_count; i++)
+            instance->phys_devs[i].obj.loader_magic = VULKAN_ICD_MAGIC_VALUE;
 
         params.pCreateInfo = create_info;
         params.pAllocator = allocator;
         params.pInstance = ret;
+        params.client_ptr = instance;
         status = UNIX_CALL(vkCreateInstance, &params);
         assert(!status);
-        if (instance->physical_device_count <= device_count)
+        if (instance->phys_dev_count <= phys_dev_count)
             break;
-        device_count = instance->physical_device_count;
+        phys_dev_count = instance->phys_dev_count;
         free(instance);
-        *ret = NULL;
     }
 
     if (params.result)
-    {
         free(instance);
-        *ret = NULL;
-    }
     return params.result;
 }
 
@@ -449,40 +393,11 @@
     free(instance);
 }
 
-static VkResult enum_host_instance_extension_properties(const char *layer_name, uint32_t *count,
-        VkExtensionProperties **properties)
-{
-    struct vkEnumerateInstanceExtensionProperties_params params =
-    {
-        .pLayerName = layer_name,
-        .pPropertyCount = count,
-    };
-    NTSTATUS status;
-
-    for (;;)
-    {
-        *properties = params.pProperties;
-        status = UNIX_CALL(vkEnumerateInstanceExtensionProperties, &params);
-        assert(!status && "vkEnumerateInstanceExtensionProperties");
-        if (params.result == VK_INCOMPLETE)
-        {
-            free(params.pProperties);
-            params.pProperties = NULL;
-            continue;
-        }
-        if (params.result || params.pProperties) return params.result;
-        params.pProperties = realloc(*properties, *params.pPropertyCount * sizeof(*params.pProperties));
-        if (!params.pProperties) return VK_ERROR_OUT_OF_HOST_MEMORY;
-    }
-}
-
 VkResult WINAPI vkEnumerateInstanceExtensionProperties(const char *layer_name,
         uint32_t *count, VkExtensionProperties *properties)
 {
-    uint32_t i, len, capacity = properties ? *count : UINT_MAX, host_count = 0;
-    struct vulkan_instance_extensions extensions = {0};
-    VkExtensionProperties *host_properties;
-    VkResult res;
+    struct vkEnumerateInstanceExtensionProperties_params params;
+    NTSTATUS status;
 
     TRACE("%p, %p, %p\n", layer_name, count, properties);
 
@@ -498,116 +413,12 @@
         return VK_SUCCESS;
     }
 
-    if ((res = enum_host_instance_extension_properties(layer_name, &host_count, &host_properties)))
-        goto done;
-
-    TRACE("Client instance extensions:\n");
-    for (i = 0, len = 0; i < host_count; i++)
-    {
-        const char *extension = host_properties[i].extensionName;
-        if (!is_instance_extension_supported(extension, &extensions)) continue;
-        TRACE("  - %s\n", extension);
-        if (len++ < capacity && properties) properties[len - 1] = host_properties[i];
-    }
-    if (instance_extensions.has_VK_KHR_win32_surface)
-    {
-        static const VkExtensionProperties VK_KHR_win32_surface = {VK_KHR_WIN32_SURFACE_EXTENSION_NAME, VK_KHR_WIN32_SURFACE_SPEC_VERSION};
-        TRACE("  - VK_KHR_win32_surface\n");
-        if (len++ < capacity && properties) properties[len - 1] = VK_KHR_win32_surface;
-    }
-    res = len > capacity ? VK_INCOMPLETE : VK_SUCCESS;
-    *count = min(len, capacity);
-
-done:
-    free(host_properties);
-    return res;
-}
-
-static VkResult enum_host_device_extension_properties(VkPhysicalDevice physical_device, const char *layer_name,
-        uint32_t *count, VkExtensionProperties **properties)
-{
-    struct vkEnumerateDeviceExtensionProperties_params params =
-    {
-        .physicalDevice = physical_device,
-        .pLayerName = layer_name,
-        .pPropertyCount = count,
-    };
-    NTSTATUS status;
-
-    for (;;)
-    {
-        *properties = params.pProperties;
-        status = UNIX_CALL(vkEnumerateDeviceExtensionProperties, &params);
-        assert(!status && "vkEnumerateDeviceExtensionProperties");
-        if (params.result == VK_INCOMPLETE)
-        {
-            free(params.pProperties);
-            params.pProperties = NULL;
-            continue;
-        }
-        if (params.result || params.pProperties) return params.result;
-        params.pProperties = realloc(*properties, *params.pPropertyCount * sizeof(*params.pProperties));
-        if (!params.pProperties) return VK_ERROR_OUT_OF_HOST_MEMORY;
-    }
-}
-
-VkResult WINAPI vkEnumerateDeviceExtensionProperties(VkPhysicalDevice physical_device, const char *layer_name,
-        uint32_t *count, VkExtensionProperties *properties)
-{
-    uint32_t i, len, capacity = properties ? *count : UINT_MAX, host_count = 0;
-    struct vulkan_device_extensions extensions = {0};
-    VkExtensionProperties *host_properties;
-    VkResult res;
-
-    TRACE("%p, %p, %p, %p\n", physical_device, layer_name, count, properties);
-
-    if (layer_name)
-    {
-        WARN("Layer enumeration not supported from ICD.\n");
-        return VK_ERROR_LAYER_NOT_PRESENT;
-    }
-
-    if ((res = enum_host_device_extension_properties(physical_device, layer_name, &host_count, &host_properties)))
-        goto done;
-
-    TRACE("Client physical device extensions:\n");
-    for (i = 0, len = 0; i < host_count; i++)
-    {
-        const char *extension = host_properties[i].extensionName;
-        if (!is_device_extension_supported(physical_device, extension, &extensions)) continue;
-        TRACE("  - %s\n", extension);
-        if (len++ < capacity && properties) properties[len - 1] = host_properties[i];
-    }
-    if (physical_device->extensions.has_VK_KHR_external_memory_win32)
-    {
-        static const VkExtensionProperties VK_KHR_external_memory_win32 = {VK_KHR_EXTERNAL_MEMORY_WIN32_EXTENSION_NAME, VK_KHR_EXTERNAL_MEMORY_WIN32_SPEC_VERSION};
-        TRACE("  - VK_KHR_external_memory_win32\n");
-        if (len++ < capacity && properties) properties[len - 1] = VK_KHR_external_memory_win32;
-    }
-    if (physical_device->extensions.has_VK_KHR_external_fence_win32)
-    {
-        static const VkExtensionProperties VK_KHR_external_fence_win32 = {VK_KHR_EXTERNAL_FENCE_WIN32_EXTENSION_NAME, VK_KHR_EXTERNAL_FENCE_WIN32_SPEC_VERSION};
-        TRACE("  - VK_KHR_external_fence_win32\n");
-        if (len++ < capacity && properties) properties[len - 1] = VK_KHR_external_fence_win32;
-    }
-    if (physical_device->extensions.has_VK_KHR_external_semaphore_win32)
-    {
-        static const VkExtensionProperties VK_KHR_external_semaphore_win32 = {VK_KHR_EXTERNAL_SEMAPHORE_WIN32_EXTENSION_NAME, VK_KHR_EXTERNAL_SEMAPHORE_WIN32_SPEC_VERSION};
-        TRACE("  - VK_KHR_external_semaphore_win32\n");
-        if (len++ < capacity && properties) properties[len - 1] = VK_KHR_external_semaphore_win32;
-    }
-    if (physical_device->extensions.has_VK_KHR_win32_keyed_mutex)
-    {
-        static const VkExtensionProperties VK_KHR_win32_keyed_mutex = {VK_KHR_WIN32_KEYED_MUTEX_EXTENSION_NAME, VK_KHR_WIN32_KEYED_MUTEX_SPEC_VERSION};
-        TRACE("  - VK_KHR_win32_keyed_mutex\n");
-        if (len++ < capacity && properties) properties[len - 1] = VK_KHR_win32_keyed_mutex;
-    }
-    res = len > capacity ? VK_INCOMPLETE : VK_SUCCESS;
-    *count = min(len, capacity);
-
-done:
-    free(host_properties);
-    return res;
+    params.pLayerName = layer_name;
+    params.pPropertyCount = count;
+    params.pProperties = properties;
+    status = UNIX_CALL(vkEnumerateInstanceExtensionProperties, &params);
+    assert(!status);
+    return params.result;
 }
 
 VkResult WINAPI vkEnumerateInstanceVersion(uint32_t *version)
@@ -629,45 +440,151 @@
     return params.result;
 }
 
-VkResult WINAPI vkCreateDevice(VkPhysicalDevice physical_device, const VkDeviceCreateInfo *create_info,
+static HANDLE get_display_device_init_mutex(void)
+{
+    HANDLE mutex = CreateMutexW(NULL, FALSE, L"display_device_init");
+
+    WaitForSingleObject(mutex, INFINITE);
+    return mutex;
+}
+
+static void release_display_device_init_mutex(HANDLE mutex)
+{
+    ReleaseMutex(mutex);
+    CloseHandle(mutex);
+}
+
+/* Wait until graphics driver is loaded by explorer */
+static void wait_graphics_driver_ready(void)
+{
+    static BOOL ready = FALSE;
+
+    if (!ready)
+    {
+        SendMessageW(GetDesktopWindow(), WM_NULL, 0, 0);
+        ready = TRUE;
+    }
+}
+
+static void fill_luid_property(VkPhysicalDeviceProperties2 *properties2)
+{
+    VkPhysicalDeviceVulkan11Properties *vk11;
+    VkBool32 device_luid_valid = VK_FALSE;
+    VkPhysicalDeviceIDProperties *id;
+    uint32_t device_node_mask = 0;
+    SP_DEVINFO_DATA device_data;
+    const uint8_t* device_uuid;
+    DWORD type, device_idx = 0;
+    HDEVINFO devinfo;
+    HANDLE mutex;
+    GUID uuid;
+    LUID luid;
+
+    vk11 = wine_vk_find_struct(properties2, PHYSICAL_DEVICE_VULKAN_1_1_PROPERTIES);
+    id = wine_vk_find_struct(properties2, PHYSICAL_DEVICE_ID_PROPERTIES);
+
+    if (!vk11 && !id)
+        return;
+
+    wait_graphics_driver_ready();
+    mutex = get_display_device_init_mutex();
+    devinfo = SetupDiGetClassDevsW(&GUID_DEVCLASS_DISPLAY, L"PCI", NULL, 0);
+    device_data.cbSize = sizeof(device_data);
+    while (SetupDiEnumDeviceInfo(devinfo, device_idx++, &device_data))
+    {
+        if (!SetupDiGetDevicePropertyW(devinfo, &device_data, &WINE_DEVPROPKEY_GPU_VULKAN_UUID,
+                &type, (BYTE *)&uuid, sizeof(uuid), NULL, 0))
+            continue;
+
+        device_uuid = id ? id->deviceUUID : vk11->deviceUUID;
+
+        if (!IsEqualGUID(&uuid, device_uuid))
+            continue;
+
+        if (SetupDiGetDevicePropertyW(devinfo, &device_data, &DEVPROPKEY_GPU_LUID, &type,
+                (BYTE *)&luid, sizeof(luid), NULL, 0))
+        {
+            device_luid_valid = VK_TRUE;
+            device_node_mask = 1;
+            break;
+        }
+    }
+    SetupDiDestroyDeviceInfoList(devinfo);
+    release_display_device_init_mutex(mutex);
+
+    if (id)
+    {
+        if (device_luid_valid) memcpy(&id->deviceLUID, &luid, sizeof(id->deviceLUID));
+        id->deviceLUIDValid = device_luid_valid;
+        id->deviceNodeMask = device_node_mask;
+    }
+
+    if (vk11)
+    {
+        if (device_luid_valid) memcpy(&vk11->deviceLUID, &luid, sizeof(vk11->deviceLUID));
+        vk11->deviceLUIDValid = device_luid_valid;
+        vk11->deviceNodeMask = device_node_mask;
+    }
+
+    TRACE("deviceName:%s deviceLUIDValid:%d LUID:%08lx:%08lx deviceNodeMask:%#x.\n",
+            properties2->properties.deviceName, device_luid_valid, luid.HighPart, luid.LowPart,
+            device_node_mask);
+}
+
+void WINAPI vkGetPhysicalDeviceProperties2(VkPhysicalDevice phys_dev,
+        VkPhysicalDeviceProperties2 *properties2)
+{
+    struct vkGetPhysicalDeviceProperties2_params params;
+    NTSTATUS status;
+
+    TRACE("%p, %p\n", phys_dev, properties2);
+
+    params.physicalDevice = phys_dev;
+    params.pProperties = properties2;
+    status = UNIX_CALL(vkGetPhysicalDeviceProperties2, &params);
+    assert(!status);
+    fill_luid_property(properties2);
+}
+
+void WINAPI vkGetPhysicalDeviceProperties2KHR(VkPhysicalDevice phys_dev,
+        VkPhysicalDeviceProperties2 *properties2)
+{
+    struct vkGetPhysicalDeviceProperties2KHR_params params;
+    NTSTATUS status;
+
+    TRACE("%p, %p\n", phys_dev, properties2);
+
+    params.physicalDevice = phys_dev;
+    params.pProperties = properties2;
+    status = UNIX_CALL(vkGetPhysicalDeviceProperties2KHR, &params);
+    assert(!status);
+    fill_luid_property(properties2);
+}
+
+VkResult WINAPI vkCreateDevice(VkPhysicalDevice phys_dev, const VkDeviceCreateInfo *create_info,
                                const VkAllocationCallbacks *allocator, VkDevice *ret)
 {
-    struct vulkan_device_extensions extensions = {0};
     struct vkCreateDevice_params params;
     uint32_t queue_count = 0, i;
     VkDevice device;
     NTSTATUS status;
-
-    TRACE("Enabling %u client device extensions\n", create_info->enabledExtensionCount);
-    for (uint32_t i = 0; i < create_info->enabledExtensionCount; i++)
-    {
-        const char *extension = create_info->ppEnabledExtensionNames[i];
-        if (!is_device_extension_supported(physical_device, extension, &extensions))
-            return VK_ERROR_EXTENSION_NOT_PRESENT;
-        TRACE("  - %s\n", extension);
-    }
 
     for (i = 0; i < create_info->queueCreateInfoCount; i++)
         queue_count += create_info->pQueueCreateInfos[i].queueCount;
     if (!(device = vulkan_client_object_create(FIELD_OFFSET(struct VkDevice_T, queues[queue_count]))))
         return VK_ERROR_OUT_OF_HOST_MEMORY;
-    for (VkQueue queue = device->queues, end = queue + queue_count; queue < end; queue++)
-        queue->obj.loader_magic = VULKAN_ICD_MAGIC_VALUE;
-    device->extensions = extensions;
-    *ret = device;
-
-    params.physicalDevice = physical_device;
+    for (i = 0; i < queue_count; i++)
+        device->queues[i].obj.loader_magic = VULKAN_ICD_MAGIC_VALUE;
+
+    params.physicalDevice = phys_dev;
     params.pCreateInfo = create_info;
     params.pAllocator = allocator;
     params.pDevice = ret;
-
+    params.client_ptr = device;
     status = UNIX_CALL(vkCreateDevice, &params);
     assert(!status);
     if (params.result)
-    {
         free(device);
-        *ret = NULL;
-    }
     return params.result;
 }
 
