diff --git a/dlls/d3dx9_36/tests/volume.c b/dlls/d3dx9_36/tests/volume.c
--- a/dlls/d3dx9_36/tests/volume.c
+++ b/dlls/d3dx9_36/tests/volume.c
@@ -23,13 +23,6 @@
 #include "d3dx9tex.h"
 #include "d3dx9_test_images.h"
 
-#define check_release(obj, exp) _check_release(__LINE__, obj, exp)
-static inline void _check_release(unsigned int line, IUnknown *obj, int exp)
-{
-    int ref = IUnknown_Release(obj);
-    ok_(__FILE__, line)(ref == exp, "Invalid refcount. Expected %d, got %d\n", exp, ref);
-}
-
 #define check_pixel_4bpp(box, x, y, z, color) _check_pixel_4bpp(__LINE__, box, x, y, z, color)
 static inline void _check_pixel_4bpp(unsigned int line, const D3DLOCKED_BOX *box, int x, int y, int z, DWORD expected_color)
 {
@@ -42,6 +35,16 @@
 {
    uint64_t color = ((uint64_t *)box->pBits)[x + (y * box->RowPitch + z * box->SlicePitch) / 8];
    ok_(__FILE__, line)(color == expected_color, "Got color %#I64x, expected %#I64x.\n", color, expected_color);
+}
+
+static inline void set_box(D3DBOX *box, UINT left, UINT top, UINT right, UINT bottom, UINT front, UINT back)
+{
+    box->Left = left;
+    box->Top = top;
+    box->Right = right;
+    box->Bottom = bottom;
+    box->Front = front;
+    box->Back = back;
 }
 
 static void test_D3DXLoadVolumeFromMemory(IDirect3DDevice9 *device)
@@ -190,8 +193,8 @@
 
     IDirect3DVolumeTexture9_GetVolumeLevel(volume_texture, 0, &volume);
 
-    set_box(&src_box, 1, 1, 5, 3, 0, 1);
-    set_box(&dst_box, 1, 1, 5, 3, 0, 1);
+    set_box(&src_box, 1, 1, 7, 7, 0, 1);
+    set_box(&dst_box, 1, 1, 7, 7, 0, 1);
     hr = D3DXLoadVolumeFromMemory(volume, NULL, &dst_box, pixels, D3DFMT_A8R8G8B8, 16, 32, NULL, &src_box, D3DX_DEFAULT, 0);
     ok(hr == D3D_OK, "D3DXLoadVolumeFromMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
 
@@ -396,15 +399,15 @@
     v->z = z;
 }
 
-static const D3DXVECTOR4 quadrant_color[] =
-{
+static const D3DXVECTOR4 quadrant_color[] = {
     { 1.0f, 0.0f, 0.0f, 1.0f }, { 0.0f, 1.0f, 0.0f, 1.0f }, { 0.0f, 0.0f, 1.0f, 1.0f }, { 1.0f, 1.0f, 1.0f, 1.0f },
     { 1.0f, 1.0f, 1.0f, 1.0f }, { 0.0f, 0.0f, 1.0f, 1.0f }, { 1.0f, 0.0f, 0.0f, 1.0f }, { 0.0f, 1.0f, 0.0f, 1.0f },
 };
 
-static void WINAPI fill_func_volume(D3DXVECTOR4 *value, const D3DXVECTOR3 *coord, const D3DXVECTOR3 *size, void *data)
-{
-    D3DXVECTOR3 vec = *coord;
+static void WINAPI fill_func_volume(D3DXVECTOR4 *value, const D3DXVECTOR3 *texcoord,
+                                   const D3DXVECTOR3 *texelsize, void *data)
+{
+    D3DXVECTOR3 vec = *texcoord;
     uint32_t idx;
 
     if (data)
@@ -413,74 +416,54 @@
         return;
     }
 
-    set_vec3(&vec, (vec.x / size->x) - 0.5f, (vec.y / size->y) - 0.5f, (vec.z / size->z) - 0.5f);
-    if (vec.x < 16.0f)
-        idx = vec.y < 16.0f ? 0 : 2;
+    set_vec3(&vec, (vec.x / texelsize->x) - 0.5f, (vec.y / texelsize->y) - 0.5f, (vec.z / texelsize->z) - 0.5f);
+    if (vec.x < 8.0f)
+        idx = vec.y < 8.0f ? 0 : 2;
     else
-        idx = vec.y < 16.0f ? 1 : 3;
+        idx = vec.y < 8.0f ? 1 : 3;
     idx += vec.z < 1.0f ? 0 : 4;
 
     *value = quadrant_color[idx];
 }
 
-#define check_volume_readback_slice_quadrants(rb, width, height, slice, expected, max_diff) \
-    _check_volume_readback_slice_quadrants(__LINE__, rb, width, height, slice, expected, max_diff)
-static inline void _check_volume_readback_slice_quadrants(unsigned int line, struct volume_readback *rb,
-        uint32_t width, uint32_t height, uint32_t slice, const uint32_t *expected, uint8_t max_diff)
-{
-    /* Order is: top left, top right, bottom left, bottom right. */
-    _check_volume_readback_pixel_4bpp_diff(__FILE__, line, rb, 0, 0, slice, expected[0], max_diff, FALSE);
-    _check_volume_readback_pixel_4bpp_diff(__FILE__, line, rb, (width - 1), 0, slice, expected[1], max_diff, FALSE);
-    _check_volume_readback_pixel_4bpp_diff(__FILE__, line, rb, 0, (height - 1), slice, expected[2], max_diff, FALSE);
-    _check_volume_readback_pixel_4bpp_diff(__FILE__, line, rb, (width - 1), (height - 1), slice, expected[3], max_diff,
-            FALSE);
-}
-
 static void test_d3dx_save_volume_to_file(IDirect3DDevice9 *device)
 {
+    static const struct
+    {
+        D3DXIMAGE_FILEFORMAT file_format;
+        const char *file_name_a;
+        const WCHAR *file_name_w;
+        D3DFORMAT save_format;
+    } save_files[] = {
+        { D3DXIFF_BMP, "saved_volume_a.bmp", L"saved_volume_w.bmp", D3DFMT_A8R8G8B8 },
+        { D3DXIFF_JPG, "saved_volume_a.jpg", L"saved_volume_w.jpg", D3DFMT_X8R8G8B8 },
+        { D3DXIFF_TGA, "saved_volume_a.tga", L"saved_volume_w.tga", D3DFMT_A8R8G8B8 },
+        { D3DXIFF_PNG, "saved_volume_a.png", L"saved_volume_w.png", D3DFMT_A8R8G8B8 },
+        { D3DXIFF_DIB, "saved_volume_a.dib", L"saved_volume_w.dib", D3DFMT_A8R8G8B8 },
+        { D3DXIFF_DDS, "saved_volume_a.dds", L"saved_volume_w.dds", D3DFMT_A8R8G8B8 },
+    };
     static const struct dds_pixel_format d3dfmt_a8r8g8b8_pf = { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32,
                                                                 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 };
-    static const uint32_t front_expected[] = { 0xffff0000, 0xff00ff00, 0xff0000ff, 0xffffffff };
-    static const uint32_t empty_expected[] = { 0x00000000, 0x00000000, 0x00000000, 0x00000000 };
-    static const uint32_t back_expected[] = { 0xffffffff, 0xff0000ff, 0xffff0000, 0xff00ff00 };
-    static const struct
-    {
-        const char *name;
-
-        D3DXIMAGE_FILEFORMAT iff;
-        D3DFORMAT saved_format;
-        uint8_t max_diff;
-        BOOL wine_todo;
-    } tests[] =
-    {
-        { "D3DXIFF_BMP", D3DXIFF_BMP, D3DFMT_A8R8G8B8 },
-        { "D3DXIFF_JPG", D3DXIFF_JPG, D3DFMT_X8R8G8B8, .max_diff = 1 },
-        { "D3DXIFF_TGA", D3DXIFF_TGA, D3DFMT_A8R8G8B8 },
-        { "D3DXIFF_PNG", D3DXIFF_PNG, D3DFMT_A8R8G8B8 },
-        { "D3DXIFF_DDS", D3DXIFF_DDS, D3DFMT_A8R8G8B8 },
-        { "D3DXIFF_HDR", D3DXIFF_HDR, D3DFMT_A32B32G32R32F, .max_diff = 1, .wine_todo = TRUE },
-        { "D3DXIFF_PFM", D3DXIFF_PFM, D3DFMT_A32B32G32R32F, .wine_todo = TRUE },
-        { "D3DXIFF_PPM", D3DXIFF_PPM, D3DFMT_X8R8G8B8, .wine_todo = TRUE },
-    };
     struct
     {
          DWORD magic;
          struct dds_header header;
          BYTE *data;
     } *dds;
+    static const uint32_t front_expected[] = { 0xffff0000, 0xff00ff00, 0xff0000ff, 0xffffffff };
+    static const uint32_t back_expected[] = { 0xffffffff, 0xff0000ff, 0xffff0000, 0xff00ff00 };
+    static const uint32_t coords[][2] = { { 0, 0 }, { 15, 0 }, { 0, 15 }, { 15, 15 } };
     const D3DXVECTOR4 clear_val = { 0.0f, 0.0f, 0.0f, 0.0f };
     IDirect3DVolumeTexture9 *volume_texture;
     struct volume_readback volume_rb;
     ID3DXBuffer *buffer = NULL;
-    WCHAR name_buf_w[MAX_PATH];
-    char name_buf_a[MAX_PATH];
     IDirect3DVolume9 *volume;
     D3DXIMAGE_INFO info;
-    unsigned int i, j;
+    uint32_t i, j, k;
     D3DBOX box;
     HRESULT hr;
 
-    hr = IDirect3DDevice9_CreateVolumeTexture(device, 32, 32, 2, 1, D3DUSAGE_DYNAMIC, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT,
+    hr = IDirect3DDevice9_CreateVolumeTexture(device, 16, 16, 2, 1, D3DUSAGE_DYNAMIC, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT,
             &volume_texture, NULL);
     if (FAILED(hr))
     {
@@ -493,160 +476,157 @@
 
     IDirect3DVolumeTexture9_GetVolumeLevel(volume_texture, 0, &volume);
 
+    set_box(&box, 0, 0, 16, 16, 1, 2);
     hr = D3DXSaveVolumeToFileInMemory(&buffer, D3DXIFF_DDS, volume, NULL, NULL);
     ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
 
     dds = ID3DXBuffer_GetBufferPointer(buffer);
-    check_dds_header(&dds->header, DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_DEPTH | DDSD_PIXELFORMAT, 32, 32, 0,
+    check_dds_header(&dds->header, DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_DEPTH | DDSD_PIXELFORMAT, 16, 16, 0,
             2, 0, &d3dfmt_a8r8g8b8_pf, DDSCAPS_TEXTURE | DDSCAPS_ALPHA, DDSCAPS2_VOLUME, FALSE);
     ID3DXBuffer_Release(buffer);
 
     /*
-     * Save with a box that has a depth of 1. This results in a DDS header
-     * without any depth/volume flags set, even though we're saving a volume.
+     * Box only has a depth of 1, saves like a regular surface. E.g no depth
+     * flags/fields set.
      */
-    set_box(&box, 0, 0, 32, 32, 1, 2);
+    set_box(&box, 0, 0, 16, 16, 1, 2);
     hr = D3DXSaveVolumeToFileInMemory(&buffer, D3DXIFF_DDS, volume, NULL, &box);
     ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
 
     dds = ID3DXBuffer_GetBufferPointer(buffer);
-    check_dds_header(&dds->header, DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT, 32, 32, 0, 0, 0,
+    check_dds_header(&dds->header, DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT, 16, 16, 0, 0, 0,
             &d3dfmt_a8r8g8b8_pf, DDSCAPS_TEXTURE | DDSCAPS_ALPHA, 0, FALSE);
     ID3DXBuffer_Release(buffer);
 
-    GetTempPathA(ARRAY_SIZE(name_buf_a), name_buf_a);
-    GetTempFileNameA(name_buf_a, "dxa", 0, name_buf_a);
-
-    GetTempPathW(ARRAY_SIZE(name_buf_w), name_buf_w);
-    GetTempFileNameW(name_buf_w, L"dxw", 0, name_buf_w);
-    for (i = 0; i < ARRAY_SIZE(tests); ++i)
-    {
-        winetest_push_context("File format %u (%s)", i, tests[i].name);
-
-        /*
-         * Test twice, first with a box that has a depth of 2, then second
-         * with a box that has a depth of 1.
-         */
+    for (i = 0; i < ARRAY_SIZE(save_files); ++i)
+    {
+        winetest_push_context("Test %u", i);
+
+        /* ASCII string. */
         for (j = 0; j < 2; ++j)
         {
-            winetest_push_context("Depth %u", 2 - j);
+            const uint32_t *expected_colors = !j ? front_expected : back_expected;
+
             hr = D3DXFillVolumeTexture(volume_texture, fill_func_volume, NULL);
             ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
 
-            set_box(&box, 0, 0, 32, 32, j, 2);
-
-            hr = D3DXSaveVolumeToFileA(name_buf_a, tests[i].iff, volume, NULL, &box);
-            todo_wine_if(tests[i].wine_todo) ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-            hr = D3DXSaveVolumeToFileW(name_buf_w, tests[i].iff, volume, NULL, &box);
-            todo_wine_if(tests[i].wine_todo) ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-            hr = D3DXSaveVolumeToFileInMemory(&buffer, tests[i].iff, volume, NULL, &box);
-            todo_wine_if(tests[i].wine_todo) ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-            if (!strcmp(winetest_platform, "wine") && tests[i].wine_todo)
-            {
-                winetest_pop_context();
-                continue;
-            }
-
-            /* ASCII string. */
+            hr = D3DXSaveVolumeToFileA(save_files[i].file_name_a, save_files[i].file_format, volume, NULL, !j ? NULL : &box);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
             hr = D3DXFillVolumeTexture(volume_texture, fill_func_volume, (void *)&clear_val);
             ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
 
             memset(&info, 0, sizeof(info));
-            hr = D3DXLoadVolumeFromFileA(volume, NULL, NULL, name_buf_a, NULL, D3DX_FILTER_NONE, 0, &info);
-            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+            hr = D3DXLoadVolumeFromFileA(volume, NULL, NULL, save_files[i].file_name_a, NULL, D3DX_FILTER_NONE, 0, &info);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
             get_texture_volume_readback(device, volume_texture, 0, &volume_rb);
-            if (tests[i].iff == D3DXIFF_DDS)
-            {
-                if (!j)
-                {
-                    check_image_info(&info, 32, 32, 2, 1, D3DFMT_A8R8G8B8, D3DRTYPE_VOLUMETEXTURE, D3DXIFF_DDS, FALSE);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, front_expected, 0);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, back_expected, 0);
-                }
-                else
-                {
-                    check_image_info(&info, 32, 32, 1, 1, D3DFMT_A8R8G8B8, D3DRTYPE_TEXTURE, D3DXIFF_DDS, FALSE);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, back_expected, 0);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, empty_expected, 0);
-                }
+            if (save_files[i].file_format == D3DXIFF_DDS)
+            {
+                check_image_info(&info, 16, 16, !j ? 2 : 1, 1, save_files[i].save_format, !j ? D3DRTYPE_VOLUMETEXTURE : D3DRTYPE_TEXTURE,
+                        D3DXIFF_DDS, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0],  coords[k][1], 0, expected_colors[k], 4, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0],  coords[k][1], 1, !j ? back_expected[k] : 0, 4, FALSE);
             }
             else
             {
-                check_image_info(&info, 32, 32, 1, 1, tests[i].saved_format, D3DRTYPE_TEXTURE, tests[i].iff, FALSE);
-                check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, !j ? front_expected : back_expected,
-                        tests[i].max_diff);
-                check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, empty_expected, tests[i].max_diff);
+                const D3DXIMAGE_FILEFORMAT iff = save_files[i].file_format == D3DXIFF_DIB ? D3DXIFF_BMP : save_files[i].file_format;
+                const uint8_t max_diff = iff == D3DXIFF_JPG ? 40 : 0;
+
+                check_image_info(&info, 16, 16, 1, 1, save_files[i].save_format, D3DRTYPE_TEXTURE, iff, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0],  coords[k][1], 0, expected_colors[k], max_diff, FALSE);
             }
             release_volume_readback(&volume_rb);
 
-            /* Wide string. */
+            DeleteFileA(save_files[i].file_name_a);
+        }
+
+        /* Wide string. */
+        for (j = 0; j < 2; ++j)
+        {
+            const uint32_t *expected_colors = !j ? front_expected : back_expected;
+
+            hr = D3DXFillVolumeTexture(volume_texture, fill_func_volume, NULL);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+            hr = D3DXSaveVolumeToFileW(save_files[i].file_name_w, save_files[i].file_format, volume, NULL, !j ? NULL : &box);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
             hr = D3DXFillVolumeTexture(volume_texture, fill_func_volume, (void *)&clear_val);
             ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
 
             memset(&info, 0, sizeof(info));
-            hr = D3DXLoadVolumeFromFileW(volume, NULL, NULL, name_buf_w, NULL, D3DX_FILTER_NONE, 0, &info);
-            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+            hr = D3DXLoadVolumeFromFileW(volume, NULL, NULL, save_files[i].file_name_w, NULL, D3DX_FILTER_NONE, 0, &info);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
             get_texture_volume_readback(device, volume_texture, 0, &volume_rb);
-            if (tests[i].iff == D3DXIFF_DDS)
-            {
-                if (!j)
-                {
-                    check_image_info(&info, 32, 32, 2, 1, D3DFMT_A8R8G8B8, D3DRTYPE_VOLUMETEXTURE, D3DXIFF_DDS, FALSE);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, front_expected, 0);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, back_expected, 0);
-                }
-                else
-                {
-                    check_image_info(&info, 32, 32, 1, 1, D3DFMT_A8R8G8B8, D3DRTYPE_TEXTURE, D3DXIFF_DDS, FALSE);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, back_expected, 0);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, empty_expected, 0);
-                }
+            if (save_files[i].file_format == D3DXIFF_DDS)
+            {
+                check_image_info(&info, 16, 16, !j ? 2 : 1, 1, save_files[i].save_format, !j ? D3DRTYPE_VOLUMETEXTURE : D3DRTYPE_TEXTURE,
+                        D3DXIFF_DDS, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0], coords[k][1], 0, expected_colors[k], 0, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0], coords[k][1], 1, !j ? back_expected[k] : 0, 0, FALSE);
             }
             else
             {
-                check_image_info(&info, 32, 32, 1, 1, tests[i].saved_format, D3DRTYPE_TEXTURE, tests[i].iff, FALSE);
-                check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, !j ? front_expected : back_expected,
-                        tests[i].max_diff);
-                check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, empty_expected, tests[i].max_diff);
+                const D3DXIMAGE_FILEFORMAT iff = save_files[i].file_format == D3DXIFF_DIB ? D3DXIFF_BMP : save_files[i].file_format;
+                const uint8_t max_diff = iff == D3DXIFF_JPG ? 40 : 0;
+
+                check_image_info(&info, 16, 16, 1, 1, save_files[i].save_format, D3DRTYPE_TEXTURE, iff, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0],  coords[k][1], 0, expected_colors[k], max_diff, FALSE);
             }
             release_volume_readback(&volume_rb);
 
-            /* File in memory. */
+            DeleteFileW(save_files[i].file_name_w);
+        }
+
+        /* InMemory. */
+        for (j = 0; j < 2; ++j)
+        {
+            const uint32_t *expected_colors = !j ? front_expected : back_expected;
+            ID3DXBuffer *buffer = NULL;
+
+            hr = D3DXFillVolumeTexture(volume_texture, fill_func_volume, NULL);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+            hr = D3DXSaveVolumeToFileInMemory(&buffer, save_files[i].file_format, volume, NULL, !j ? NULL : &box);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+            hr = D3DXFillVolumeTexture(volume_texture, fill_func_volume, (void *)&clear_val);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
             memset(&info, 0, sizeof(info));
             hr = D3DXLoadVolumeFromFileInMemory(volume, NULL, NULL, ID3DXBuffer_GetBufferPointer(buffer),
                     ID3DXBuffer_GetBufferSize(buffer), NULL, D3DX_FILTER_NONE, 0, &info);
             ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+            ID3DXBuffer_Release(buffer);
+
             get_texture_volume_readback(device, volume_texture, 0, &volume_rb);
-            if (tests[i].iff == D3DXIFF_DDS)
-            {
-                if (!j)
-                {
-                    check_image_info(&info, 32, 32, 2, 1, D3DFMT_A8R8G8B8, D3DRTYPE_VOLUMETEXTURE, D3DXIFF_DDS, FALSE);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, front_expected, 0);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, back_expected, 0);
-                }
-                else
-                {
-                    check_image_info(&info, 32, 32, 1, 1, D3DFMT_A8R8G8B8, D3DRTYPE_TEXTURE, D3DXIFF_DDS, FALSE);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, back_expected, 0);
-                    check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, empty_expected, 0);
-                }
+            if (save_files[i].file_format == D3DXIFF_DDS)
+            {
+                check_image_info(&info, 16, 16, !j ? 2 : 1, 1, save_files[i].save_format, !j ? D3DRTYPE_VOLUMETEXTURE : D3DRTYPE_TEXTURE,
+                        D3DXIFF_DDS, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0],  coords[k][1], 0, expected_colors[k], 4, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0],  coords[k][1], 1, !j ? back_expected[k] : 0, 4, FALSE);
             }
             else
             {
-                check_image_info(&info, 32, 32, 1, 1, tests[i].saved_format, D3DRTYPE_TEXTURE, tests[i].iff, FALSE);
-                check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 0, !j ? front_expected : back_expected,
-                        tests[i].max_diff);
-                check_volume_readback_slice_quadrants(&volume_rb, 32, 32, 1, empty_expected, tests[i].max_diff);
+                const D3DXIMAGE_FILEFORMAT iff = save_files[i].file_format == D3DXIFF_DIB ? D3DXIFF_BMP : save_files[i].file_format;
+                const uint8_t max_diff = iff == D3DXIFF_JPG ? 40 : 0;
+
+                check_image_info(&info, 16, 16, 1, 1, save_files[i].save_format, D3DRTYPE_TEXTURE, iff, FALSE);
+                for (k = 0; k < ARRAY_SIZE(coords); ++k)
+                    check_volume_readback_pixel_4bpp_diff(&volume_rb, coords[k][0],  coords[k][1], 0, expected_colors[k], max_diff, FALSE);
             }
             release_volume_readback(&volume_rb);
-
-            ID3DXBuffer_Release(buffer);
-            DeleteFileA(name_buf_a);
-            DeleteFileW(name_buf_w);
-            winetest_pop_context();
         }
 
         winetest_pop_context();
@@ -654,125 +634,6 @@
 
     IDirect3DVolume9_Release(volume);
     IDirect3DVolumeTexture9_Release(volume_texture);
-}
-
-static void test_image_filters(void)
-{
-    static const struct
-    {
-        DWORD filter;
-        const char *name;
-    }
-    test_filters[] =
-    {
-        { D3DX_FILTER_POINT, "D3DX_FILTER_POINT" },
-        { D3DX_FILTER_LINEAR, "D3DX_FILTER_LINEAR" },
-        { D3DX_FILTER_TRIANGLE, "D3DX_FILTER_TRIANGLE" },
-        { D3DX_FILTER_BOX, "D3DX_FILTER_BOX" },
-    };
-    static const struct
-    {
-        const D3DBOX src_box;
-        const void *src_data;
-        D3DFORMAT format;
-
-        unsigned int dst_width;
-        unsigned int dst_height;
-        unsigned int dst_depth;
-        struct
-        {
-            const void *expected_dst_data;
-            const void *expected_dst_data_32;
-            BOOL todo;
-            uint32_t max_diff;
-        }
-        filter_expected[4];
-    } tests[] =
-    {
-        {
-            { 0, 0, 8, 8, 0, 8 }, a8r8g8b8_8_8_8, D3DFMT_A8R8G8B8, 4, 4, 4,
-            {
-                { a8r8g8b8_8_8_8_point_filter_4_4_4 },
-                { a8r8g8b8_8_8_8_linear_filter_4_4_4, .todo = TRUE },
-                {
-                    a8r8g8b8_8_8_8_triangle_filter_4_4_4, a8r8g8b8_8_8_8_triangle_filter_4_4_4_32bit,
-                    .todo = TRUE
-                },
-                /* Linear and box filters match on 64bit, but not 32bit. */
-                {
-                    a8r8g8b8_8_8_8_linear_filter_4_4_4, a8r8g8b8_8_8_8_box_filter_4_4_4_32bit,
-#ifndef _WIN64
-                    /* We match exactly on 64-bit, but not 32-bit. */
-                    .max_diff = 1
-#endif
-                },
-            },
-        },
-        {
-            { 0, 0, 6, 6, 0, 6 }, a8r8g8b8_6_6_6, D3DFMT_A8R8G8B8, 3, 3, 3,
-            {
-                { a8r8g8b8_6_6_6_point_filter_3_3_3 },
-                { a8r8g8b8_6_6_6_linear_filter_3_3_3, .todo = TRUE },
-                { a8r8g8b8_6_6_6_triangle_filter_3_3_3, .todo = TRUE },
-                { a8r8g8b8_6_6_6_linear_filter_3_3_3, .todo = TRUE },
-            },
-        },
-    };
-    IDirect3DVolumeTexture9 *volume_texture;
-    uint32_t src_pitch, src_slice_pitch;
-    const uint8_t *expected_dst;
-    D3DLOCKED_BOX locked_box;
-    IDirect3DVolume9 *volume;
-    IDirect3DDevice9 *device;
-    unsigned int i, j;
-    HRESULT hr;
-    HWND hwnd;
-
-    if (!(device = create_device(&hwnd)))
-        return;
-
-    for (i = 0; i < ARRAY_SIZE(tests); ++i)
-    {
-        const unsigned int fmt_bpp = get_bpp_for_d3dformat(tests[i].format);
-
-        winetest_push_context("Test %u", i);
-
-        hr = IDirect3DDevice9_CreateVolumeTexture(device, tests[i].dst_width, tests[i].dst_height, tests[i].dst_depth, 1,
-                0, tests[i].format, D3DPOOL_SCRATCH, &volume_texture, NULL);
-        ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-        IDirect3DVolumeTexture9_GetVolumeLevel(volume_texture, 0, &volume);
-
-        src_pitch = fmt_bpp * (tests[i].src_box.Right - tests[i].src_box.Left);
-        src_slice_pitch = src_pitch * (tests[i].src_box.Back - tests[i].src_box.Front);
-        for (j = 0; j < ARRAY_SIZE(test_filters); ++j)
-        {
-            winetest_push_context("Filter %s", test_filters[j].name);
-
-            hr = D3DXLoadVolumeFromMemory(volume, NULL, NULL, tests[i].src_data, tests[i].format, src_pitch,
-                    src_slice_pitch, NULL, &tests[i].src_box, test_filters[j].filter, 0);
-            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-            if (sizeof(void *) == 4 && tests[i].filter_expected[j].expected_dst_data_32)
-                expected_dst = tests[i].filter_expected[j].expected_dst_data_32;
-            else
-                expected_dst = tests[i].filter_expected[j].expected_dst_data;
-
-            IDirect3DVolume9_LockBox(volume, &locked_box, NULL, D3DLOCK_READONLY);
-            todo_wine_if(tests[i].filter_expected[j].todo) check_test_readback(locked_box.pBits, locked_box.RowPitch,
-                    locked_box.SlicePitch, expected_dst, tests[i].dst_width, tests[i].dst_height, tests[i].dst_depth,
-                    tests[i].format, tests[i].filter_expected[j].max_diff);
-            IDirect3DVolume9_UnlockBox(volume);
-
-            winetest_pop_context();
-        }
-
-        check_release((IUnknown *)volume_texture, 1);
-        check_release((IUnknown *)volume, 0);
-        winetest_pop_context();
-    }
-
-    check_release((IUnknown *)device, 0);
-    DestroyWindow(hwnd);
 }
 
 START_TEST(volume)
@@ -816,6 +677,4 @@
     IDirect3DDevice9_Release(device);
     IDirect3D9_Release(d3d);
     DestroyWindow(wnd);
-
-    test_image_filters();
-}
+}
