diff --git a/dlls/winex11.drv/event.c b/dlls/winex11.drv/event.c
--- a/dlls/winex11.drv/event.c
+++ b/dlls/winex11.drv/event.c
@@ -258,6 +258,7 @@
     xev.xclient.data.l[4] = 0;
 
     XSendEvent(display, window, False, NoEventMask, &xev);
+    XFlush( display );
 }
 
 /***********************************************************************
@@ -321,6 +322,30 @@
     }
 }
 
+static void wait_grab_pointer( Display *display )
+{
+    RECT rect;
+
+    /* unnecessary on gamescope, windows cannot be moved with the mouse */
+    if (X11DRV_HasWindowManager( "steamcompmgr" )) return;
+
+    /* release cursor grab held by any Wine process */
+    NtUserGetClipCursor( &rect );
+    NtUserClipCursor( NULL );
+
+    while (XGrabPointer( display, root_window, False, 0, GrabModeAsync, GrabModeAsync,
+                         None, None, CurrentTime ) != GrabSuccess)
+    {
+        LARGE_INTEGER timeout = {.QuadPart = -10 * (ULONGLONG)10000};
+        NtDelayExecution( FALSE, &timeout );
+    }
+
+    XUngrabPointer( display, CurrentTime );
+    XFlush( display );
+
+    /* restore the previously used clipping rect */
+    NtUserClipCursor( &rect );
+}
 
 enum event_merge_action
 {
@@ -377,6 +402,10 @@
  */
 static enum event_merge_action merge_events( XEvent *prev, XEvent *next )
 {
+#ifdef HAVE_X11_EXTENSIONS_XINPUT2_H
+    struct x11drv_thread_data *thread_data = x11drv_thread_data();
+#endif
+
     switch (prev->type)
     {
     case ConfigureNotify:
@@ -408,19 +437,21 @@
         case GenericEvent:
             if (next->xcookie.extension != xinput2_opcode) break;
             if (next->xcookie.evtype != XI_RawMotion) break;
-            if (x11drv_thread_data()->warp_serial) break;
+            if (thread_data->xinput2_rawinput) break;
+            if (thread_data->warp_serial) break;
             return MERGE_KEEP;
         }
         break;
     case GenericEvent:
         if (prev->xcookie.extension != xinput2_opcode) break;
         if (prev->xcookie.evtype != XI_RawMotion) break;
+        if (thread_data->xinput2_rawinput) break;
         switch (next->type)
         {
         case GenericEvent:
             if (next->xcookie.extension != xinput2_opcode) break;
             if (next->xcookie.evtype != XI_RawMotion) break;
-            if (x11drv_thread_data()->warp_serial) break;
+            if (thread_data->warp_serial) break;
             return merge_raw_motion_events( prev->xcookie.data, next->xcookie.data );
 #endif
         }
@@ -463,45 +494,63 @@
     return ret;
 }
 
-static int check_fd_events( int fd, int events )
-{
-    struct pollfd pfd = {.fd = fd, .events = events};
-    if (poll( &pfd, 1, 0 ) <= 0) return 0;
-    return pfd.revents;
-}
 
 /***********************************************************************
- *           ProcessEvents   (X11DRV.@)
- */
-BOOL X11DRV_ProcessEvents( DWORD mask )
-{
-    struct x11drv_thread_data *data = x11drv_thread_data();
+ *           process_events
+ */
+static BOOL process_events( Display *display, Bool (*filter)(Display*, XEvent*,XPointer), ULONG_PTR arg )
+{
     XEvent event, prev_event;
     int count = 0;
+    BOOL queued = FALSE, overlay_enabled = FALSE;
     enum event_merge_action action = MERGE_DISCARD;
-
-    if (!data) return FALSE;
-    if (data->current_event) mask = 0;  /* don't process nested events */
+    ULONG_PTR overlay_filter = QS_KEY | QS_MOUSEBUTTON | QS_MOUSEMOVE;
+    LARGE_INTEGER timeout = {0};
+
+    if (NtWaitForSingleObject(steam_overlay_event, FALSE, &timeout) == WAIT_OBJECT_0)
+        overlay_enabled = TRUE;
 
     prev_event.type = 0;
-    for (;;)
-    {
-        if (!XCheckIfEvent( data->display, &event, filter_event, (XPointer)(UINT_PTR)mask ))
-        {
-            if (!prev_event.type) break;
-            call_event_handler( data->display, &prev_event );
-            free_event_data( &prev_event );
-
-            /* Retry after processing delayed event, more events might have been read from the pipe,
-             * this is the case for instance with synchronous requests like when reading a property.
+    while (XCheckIfEvent( display, &event, filter, (char *)arg ))
+    {
+        count++;
+        if (overlay_enabled && filter_event( display, &event, (char *)overlay_filter ))
+        {
+            get_event_data( &event );
+            free_event_data( &event );
+            continue;
+        }
+        if (XFilterEvent( &event, None ))
+        {
+            /*
+             * SCIM on linux filters key events strangely. It does not filter the
+             * KeyPress events for these keys however it does filter the
+             * KeyRelease events. This causes wine to become very confused as
+             * to the keyboard state.
+             *
+             * We need to let those KeyRelease events be processed so that the
+             * keyboard state is correct.
              */
-            action = MERGE_DISCARD;
-            prev_event.type = 0;
-            continue;
-        }
-
-        count++;
-        if (XFilterEvent( &event, None )) continue;
+            if (event.type == KeyRelease)
+            {
+                KeySym keysym = 0;
+                XKeyEvent *keyevent = &event.xkey;
+
+                XLookupString(keyevent, NULL, 0, &keysym, NULL);
+                if (!(keysym == XK_Shift_L ||
+                    keysym == XK_Shift_R ||
+                    keysym == XK_Control_L ||
+                    keysym == XK_Control_R ||
+                    keysym == XK_Alt_R ||
+                    keysym == XK_Alt_L ||
+                    keysym == XK_Meta_R ||
+                    keysym == XK_Meta_L))
+                        continue; /* not a key we care about, ignore it */
+            }
+            else
+                continue;  /* filtered, ignore it */
+        }
+
         if (host_window_filter_event( &event, &prev_event )) continue;
 
         get_event_data( &event );
@@ -509,29 +558,40 @@
         switch( action )
         {
         case MERGE_HANDLE:  /* handle prev, keep new */
-            call_event_handler( data->display, &prev_event );
+            queued |= call_event_handler( display, &prev_event );
             /* fall through */
         case MERGE_DISCARD:  /* discard prev, keep new */
             free_event_data( &prev_event );
             prev_event = event;
             break;
         case MERGE_KEEP:  /* handle new, keep prev for future merging */
-            call_event_handler( data->display, &event );
+            queued |= call_event_handler( display, &event );
             /* fall through */
         case MERGE_IGNORE: /* ignore new, keep prev for future merging */
             free_event_data( &event );
             break;
         }
     }
-
+    if (prev_event.type) queued |= call_event_handler( display, &prev_event );
+    free_event_data( &prev_event );
     XFlush( gdi_display );
-    if (count) TRACE( "processed %d events\n", count );
-
-    if (mask != QS_ALLINPUT || check_fd_events( ConnectionNumber( data->display ), POLLIN )) return FALSE;
-    XFlush( data->display ); /* all events have been processed, flush any pending request */
-    return TRUE;
-}
-
+    if (count) TRACE( "processed %d events, returning %d\n", count, queued );
+    return queued;
+}
+
+
+/***********************************************************************
+ *           ProcessEvents   (X11DRV.@)
+ */
+BOOL X11DRV_ProcessEvents( DWORD mask )
+{
+    struct x11drv_thread_data *data = x11drv_thread_data();
+
+    if (!data) return FALSE;
+    if (data->current_event) mask = 0;  /* don't process nested events */
+
+    return process_events( data->display, filter_event, mask );
+}
 
 /***********************************************************************
  *           EVENT_x11_time_to_win32_time
@@ -593,7 +653,6 @@
     if (style & WS_MINIMIZE) return FALSE;
     if (NtUserGetWindowLongW( hwnd, GWL_EXSTYLE ) & WS_EX_NOACTIVATE) return FALSE;
     if (hwnd == NtUserGetDesktopWindow()) return FALSE;
-    if (NtUserGetPresentRect( hwnd, &rect, 0 )) return TRUE;
     if (NtUserGetWindowRect( hwnd, &rect, NtUserGetDpiForWindow( hwnd ) ) && IsRectEmpty( &rect )) return FALSE;
     return !(style & WS_DISABLED);
 }
@@ -637,11 +696,13 @@
     Window win;
     GUITHREADINFO threadinfo;
 
+    wait_grab_pointer( display );
+
     TRACE( "setting foreground window to %p\n", focus );
 
-    if (!is_net_supported( x11drv_atom(_NET_ACTIVE_WINDOW) ))
-    {
-        NtUserSetForegroundWindowInternal( focus );
+    if (X11DRV_HasWindowManager( "steamcompmgr" ) || !is_net_supported( x11drv_atom(_NET_ACTIVE_WINDOW) ))
+    {
+        NtUserSetForegroundWindow( focus );
 
         threadinfo.cbSize = sizeof(threadinfo);
         NtUserGetGUIThreadInfo( 0, &threadinfo );
@@ -727,7 +788,7 @@
                         NtUserSetActiveWindow( hwnd );
                         break;
                     default:
-                        WARN( "unknown WM_MOUSEACTIVATE code %ld\n", ma );
+                        WARN( "unknown WM_MOUSEACTIVATE code %d\n", (int) ma );
                         break;
                 }
             }
@@ -737,7 +798,8 @@
     }
     else if (protocol == x11drv_atom(WM_TAKE_FOCUS))
     {
-        HWND last_focus = x11drv_thread_data()->last_focus, foreground = NtUserGetForegroundWindow();
+        struct x11drv_thread_data *data = x11drv_thread_data();
+        HWND last_focus = data->last_focus, foreground = NtUserGetForegroundWindow();
 
         if (window_has_pending_wm_state( hwnd, -1 ) || (hwnd != foreground && !window_should_take_focus( foreground, event_time )))
         {
@@ -749,8 +811,14 @@
         TRACE( "window %p/%lx WM_TAKE_FOCUS serial %lu, event_time %ld, foreground %p\n", hwnd, event->window,
                event->serial, event_time, foreground );
         TRACE( "  enabled %u, visible %u, style %#x, focus %p, active %p, last %p\n",
-                NtUserIsWindowEnabled( hwnd ), NtUserIsWindowVisible( hwnd ), NtUserGetWindowLongW( hwnd, GWL_STYLE ),
+                NtUserIsWindowEnabled( hwnd ), NtUserIsWindowVisible( hwnd ), (int)NtUserGetWindowLongW( hwnd, GWL_STYLE ),
                 get_focus(), get_active_window(), last_focus );
+
+        /* Steam sometimes calls XSetInputFocus with CurrentTime when it gets focused out, and the game gets
+         * focused in, this effectively sometimes steals focus away from us. Although there's no guarantee to
+         * win the race as it entirely depends on the request timings, using CurrentTime makes it more likely.
+         */
+        if (data->active_window && !strcmp( data->active_window, "Steam" )) event_time = CurrentTime;
 
         if (can_activate_window(hwnd))
         {
@@ -842,6 +910,17 @@
     if (is_virtual_desktop() && hwnd == NtUserGetDesktopWindow()) reapply_cursor_clipping();
     if (hwnd == NtUserGetDesktopWindow()) return FALSE;
 
+    /* Focus was just restored but it can be right after super was
+     * pressed and gnome-shell needs a bit of time to respond and
+     * toggle the activity view. If we grab the cursor right away
+     * it will cancel it and super key will do nothing.
+     */
+    if (event->mode == NotifyUngrab && X11DRV_HasWindowManager( "Mutter" ))
+    {
+        LARGE_INTEGER timeout = {.QuadPart = 100 * -10000};
+        NtDelayExecution( FALSE, &timeout );
+    }
+
     x11drv_thread_data()->keymapnotify_hwnd = hwnd;
 
     /* when keyboard grab is released, re-apply the cursor clipping rect */
@@ -853,7 +932,7 @@
 
     xim_set_focus( hwnd, TRUE );
 
-    if (use_take_focus) return TRUE;
+    if (use_take_focus && !X11DRV_HasWindowManager( "steamcompmgr" )) return TRUE;
 
     if (!can_activate_window(hwnd))
     {
@@ -863,7 +942,12 @@
         if (!hwnd) hwnd = x11drv_thread_data()->last_focus;
         if (hwnd && can_activate_window(hwnd)) set_focus( event->display, hwnd, CurrentTime );
     }
-    else NtUserSetForegroundWindowInternal( hwnd );
+    else
+    {
+        wait_grab_pointer( event->display );
+        NtUserSetForegroundWindow( hwnd );
+    }
+
     return TRUE;
 }
 
@@ -885,7 +969,8 @@
     /* don't reset the foreground window, if the window which is
        getting the focus is a Wine window */
 
-    if (!is_net_supported( x11drv_atom(_NET_ACTIVE_WINDOW) ) && !is_current_process_focused())
+    if ((X11DRV_HasWindowManager( "steamcompmgr" ) || !is_net_supported( x11drv_atom(_NET_ACTIVE_WINDOW) ))
+            && !is_current_process_focused())
     {
         /* Abey : 6-Oct-99. Check again if the focus out window is the
            Foreground window, because in most cases the messages sent
@@ -894,7 +979,7 @@
         if (hwnd == NtUserGetForegroundWindow())
         {
             TRACE( "lost focus, setting fg to desktop\n" );
-            NtUserSetForegroundWindowInternal( NtUserGetDesktopWindow() );
+            NtUserSetForegroundWindow( NtUserGetDesktopWindow() );
         }
     }
  }
@@ -925,12 +1010,6 @@
     if (window_has_pending_wm_state( hwnd, NormalState )) /* ignore FocusOut only if the window is being shown */
     {
         WARN( "Ignoring window %p/%lx FocusOut serial %lu, detail %s, mode %s, foreground %p during WM_STATE change\n",
-              hwnd, event->window, event->serial, focus_details[event->detail], focus_modes[event->mode], foreground );
-        return FALSE;
-    }
-    if (!is_virtual_desktop() && window_is_reparenting( hwnd )) /* ignore FocusOut only if the window is being reparented */
-    {
-        WARN( "Ignoring window %p/%lx FocusOut serial %lu, detail %s, mode %s, foreground %p during reparenting\n",
               hwnd, event->window, event->serial, focus_details[event->detail], focus_modes[event->mode], foreground );
         return FALSE;
     }
@@ -948,6 +1027,28 @@
     return TRUE;
 }
 
+void clear_emulated_fullscreen_padding( struct x11drv_win_data *data )
+{
+    RECT rect, visible;
+
+    visible = data->rects.visible;
+    OffsetRect( &visible, -data->rects.visible.left, -data->rects.visible.top );
+
+    rect = data->rects.window;
+    OffsetRect( &rect, -data->rects.visible.left, -data->rects.visible.top );
+    intersect_rect( &rect, &rect, &visible );
+
+    if (rect.left > 0 || rect.top > 0 || rect.right < visible.right || rect.bottom < visible.bottom)
+    {
+        UINT width = visible.right - rect.right, height = visible.bottom - rect.bottom;
+
+        TRACE("clearing for visible %s, rect %s.\n", wine_dbgstr_rect(&visible), wine_dbgstr_rect(&rect));
+        if (visible.right && rect.top) XClearArea( data->display, data->whole_window, 0, 0, visible.right, rect.top, 0 );
+        if (rect.left && visible.bottom) XClearArea( data->display, data->whole_window, 0, 0, rect.left, visible.bottom, 0 );
+        if (width && visible.bottom) XClearArea( data->display, data->whole_window, rect.right, 0, width, visible.bottom, 0 );
+        if (height && visible.right) XClearArea( data->display, data->whole_window, 0, rect.bottom, visible.right, height, 0 );
+    }
+}
 
 /***********************************************************************
  *           X11DRV_Expose
@@ -958,7 +1059,7 @@
     RECT rect, abs_rect;
     POINT pos;
     struct x11drv_win_data *data;
-    UINT flags = RDW_INVALIDATE | RDW_FRAME | RDW_ALLCHILDREN;
+    UINT flags = RDW_INVALIDATE | RDW_ERASE | RDW_FRAME | RDW_ALLCHILDREN;
 
     TRACE( "win %p (%lx) %d,%d %dx%d\n",
            hwnd, event->window, event->x, event->y, event->width, event->height );
@@ -971,6 +1072,8 @@
     else pos = root_to_virtual_screen( event->x, event->y );
 
     if (!(data = get_win_data( hwnd ))) return FALSE;
+
+    clear_emulated_fullscreen_padding( data );
 
     rect.left   = pos.x;
     rect.top    = pos.y;
@@ -1013,11 +1116,6 @@
     if (event->xany.window == x11drv_thread_data()->clip_window) return TRUE;
 
     if (!(data = get_win_data( hwnd ))) return FALSE;
-    if (data->reparenting)
-    {
-        TRACE( "window %p/%lx has been reparented\n", data->hwnd, data->whole_window );
-        data->reparenting = 0;
-    }
 
     if (!data->managed && !data->embedded && data->desired_state.wm_state != WithdrawnState)
     {
@@ -1035,15 +1133,6 @@
  */
 static BOOL X11DRV_UnmapNotify( HWND hwnd, XEvent *event )
 {
-    struct x11drv_win_data *data;
-
-    if (!(data = get_win_data( hwnd ))) return FALSE;
-    if (data->managed && !data->wm_state_serial && data->current_state.wm_state == NormalState)
-    {
-        WARN( "window %p/%lx unexpectedly unmapped, assuming reparenting\n", data->hwnd, data->whole_window );
-        data->reparenting = 1;
-    }
-    release_win_data( data );
     return TRUE;
 }
 
@@ -1243,6 +1332,16 @@
 
     if (hwnd == NtUserGetForegroundWindow() && activate) set_net_active_window( hwnd, 0 );
     NtUserPostMessage( hwnd, WM_WINE_WINDOW_STATE_CHANGED, 0, 0 );
+    if (value == IconicState && X11DRV_HasWindowManager( "steamcompmgr" ) && (data = get_win_data( hwnd )))
+    {
+        if (data->whole_window)
+        {
+            FIXME( "restoring %p.\n", hwnd );
+            XUnmapWindow( data->display, data->whole_window );
+            XMapWindow( data->display, data->whole_window );
+        }
+        release_win_data( data );
+    }
 }
 
 static void handle_xembed_info_notify( HWND hwnd, XPropertyEvent *event )
@@ -1269,18 +1368,6 @@
     NtUserPostMessage( hwnd, WM_WINE_WINDOW_STATE_CHANGED, 0, 0 );
 }
 
-static void handle_wm_hints_notify( HWND hwnd, XPropertyEvent *event )
-{
-    struct x11drv_win_data *data;
-    XWMHints empty = {0}, *hints;
-
-    if (!(data = get_win_data( hwnd ))) return;
-    hints = event->state == PropertyNewValue ? XGetWMHints( event->display, event->window ) : &empty;
-    window_wm_hints_notify( data, event->serial, hints );
-    if (hints != &empty) XFree( hints );
-    release_win_data( data );
-}
-
 static void handle_mwm_hints_notify( HWND hwnd, XPropertyEvent *event )
 {
     struct x11drv_win_data *data;
@@ -1292,23 +1379,6 @@
     release_win_data( data );
 }
 
-static void handle_wm_normal_hints_notify( HWND hwnd, XPropertyEvent *event )
-{
-    struct x11drv_win_data *data;
-    XSizeHints *hints;
-    long len = 0;
-
-    if (!(data = get_win_data( hwnd ))) return;
-    if ((hints = XAllocSizeHints()))
-    {
-        if (event->state == PropertyNewValue) XGetWMNormalHints( event->display, event->window, hints, &len );
-        if (len < sizeof(*hints)) memset( (char *)hints + len, 0, sizeof(*hints) - len );
-        window_wm_normal_hints_notify( data, event->serial, hints );
-        XFree( hints );
-    }
-    release_win_data( data );
-}
-
 static void handle_net_supported_notify( XPropertyEvent *event )
 {
     struct x11drv_thread_data *data = x11drv_thread_data();
@@ -1324,9 +1394,29 @@
     if (event->state == PropertyNewValue) net_supported_init( data );
 }
 
+static void handle_net_supporting_wm_check_notify( XPropertyEvent *event )
+{
+    struct x11drv_thread_data *data = x11drv_thread_data();
+
+    if (data->window_manager)
+    {
+        XFree( data->window_manager );
+        data->window_manager = NULL;
+    }
+
+    if (event->state == PropertyNewValue) net_supporting_wm_check_init( data );
+}
+
 static void handle_net_active_window( XPropertyEvent *event )
 {
+    struct x11drv_thread_data *data = x11drv_thread_data();
     Window window = 0;
+
+    if (data->active_window)
+    {
+        XFree( data->active_window );
+        data->active_window = NULL;
+    }
 
     if (event->state == PropertyNewValue) window = get_net_active_window( event->display );
     net_active_window_notify( event->serial, window, event->time );
@@ -1343,10 +1433,9 @@
     if (event->atom == x11drv_atom(WM_STATE)) handle_wm_state_notify( hwnd, event );
     if (event->atom == x11drv_atom(_XEMBED_INFO)) handle_xembed_info_notify( hwnd, event );
     if (event->atom == x11drv_atom(_NET_WM_STATE)) handle_net_wm_state_notify( hwnd, event );
-    if (event->atom == x11drv_atom(WM_HINTS)) handle_wm_hints_notify( hwnd, event );
     if (event->atom == x11drv_atom(_MOTIF_WM_HINTS)) handle_mwm_hints_notify( hwnd, event );
-    if (event->atom == x11drv_atom(WM_NORMAL_HINTS)) handle_wm_normal_hints_notify( hwnd, event );
     if (event->atom == x11drv_atom(_NET_SUPPORTED)) handle_net_supported_notify( event );
+    if (event->atom == x11drv_atom(_NET_SUPPORTING_WM_CHECK)) handle_net_supporting_wm_check_notify( event );
     if (event->atom == x11drv_atom(_NET_ACTIVE_WINDOW)) handle_net_active_window( event );
 
     return TRUE;
@@ -1361,25 +1450,12 @@
 void X11DRV_ActivateWindow( HWND hwnd, HWND previous )
 {
     struct x11drv_win_data *data;
-    BOOL flush = FALSE;
-
-    if (!is_virtual_desktop())
-    {
-        set_net_active_window( hwnd, previous );
-        flush = TRUE;
-    }
-
-    if ((data = get_win_data( hwnd )))
-    {
-        if (!data->managed || data->embedder)
-        {
-            set_input_focus( data );
-            flush = TRUE;
-        }
-        release_win_data( data );
-    }
-
-    if (flush) XFlush( x11drv_thread_data()->display );
+
+    if (!is_virtual_desktop()) set_net_active_window( hwnd, previous );
+
+    if (!(data = get_win_data( hwnd ))) return;
+    if (!data->managed || data->embedder) set_input_focus( data );
+    release_win_data( data );
 }
 
 static void drag_drop_enter( UINT entries_size, struct format_entry *entries )
@@ -1650,7 +1726,7 @@
     effect = drag_drop_drag( hwnd, point, effect );
 
     TRACE( "actionRequested(%ld) chosen(0x%x) at x(%d),y(%d)\n",
-           event->data.l[4], effect, point.x, point.y );
+           event->data.l[4], effect, (int)point.x, (int)point.y );
 
     /*
      * Let source know if we're accepting the drop by
