diff --git a/dlls/ntdll/tests/virtual.c b/dlls/ntdll/tests/virtual.c
--- a/dlls/ntdll/tests/virtual.c
+++ b/dlls/ntdll/tests/virtual.c
@@ -51,11 +51,6 @@
 static BOOL is_wow64;
 
 static SYSTEM_BASIC_INFORMATION sbi;
-
-static inline void *get_rva( HMODULE module, DWORD va )
-{
-    return (void *)((char *)module + va);
-}
 
 static HANDLE create_target_process(const char *arg)
 {
@@ -1661,6 +1656,7 @@
         offset.QuadPart = 0;
         status = NtMapViewOfSection(mapping, process, &ptr2, 0, 0, &offset,
                                     &size, 1, AT_ROUND_TO_PAGE, PAGE_READWRITE);
+        todo_wine
         ok(status == STATUS_INVALID_PARAMETER_9 || status == STATUS_INVALID_PARAMETER,
            "NtMapViewOfSection returned %08lx\n", status);
     }
@@ -1889,7 +1885,9 @@
         size = 0;
         offset.QuadPart = 0;
         status = pNtMapViewOfSectionEx(mapping, process, &ptr2, &offset, &size, AT_ROUND_TO_PAGE, PAGE_READWRITE, NULL, 0);
-        ok(status == STATUS_INVALID_PARAMETER, "NtMapViewOfSectionEx returned %08lx\n", status);
+        todo_wine
+        ok(status == STATUS_INVALID_PARAMETER_9 || status == STATUS_INVALID_PARAMETER,
+           "NtMapViewOfSection returned %08lx\n", status);
     }
 
     status = NtUnmapViewOfSection(process, ptr);
@@ -2041,6 +2039,8 @@
     CloseHandle(process);
 }
 
+#define SUPPORTED_XSTATE_FEATURES ((1 << XSTATE_LEGACY_FLOATING_POINT) | (1 << XSTATE_LEGACY_SSE) | (1 << XSTATE_AVX))
+
 static void test_user_shared_data(void)
 {
     struct old_xstate_configuration
@@ -2052,23 +2052,21 @@
         XSTATE_FEATURE Features[MAXIMUM_XSTATE_FEATURES];
     };
 
-    ULONG feature_offsets[] =
+    static const ULONG feature_offsets[] =
     {
             0,
             160, /*offsetof(XMM_SAVE_AREA32, XmmRegisters)*/
             512  /* sizeof(XMM_SAVE_AREA32) */ + offsetof(XSTATE, YmmContext),
     };
-    ULONG feature_sizes[] =
+    static const ULONG feature_sizes[] =
     {
             160,
             256, /*sizeof(M128A) * 16 */
             sizeof(YMMCONTEXT),
     };
-    const KUSER_SHARED_DATA *user_shared_data = (void *)0x7ffe0000;
+    const KSHARED_USER_DATA *user_shared_data = (void *)0x7ffe0000;
     XSTATE_CONFIGURATION xstate = user_shared_data->XState;
     ULONG64 feature_mask;
-    ULONG64 supported_xstate_features = (1 << XSTATE_LEGACY_FLOATING_POINT) | (1 << XSTATE_LEGACY_SSE) | (1 << XSTATE_AVX);
-    ULONG xstate_part_size = sizeof(XSTATE);
     unsigned int i;
 
     ok(user_shared_data->NumberOfPhysicalPages == sbi.MmNumberOfPhysicalPages,
@@ -2110,15 +2108,6 @@
         return;
     }
 
-    if (!(xstate.EnabledFeatures & (1 << XSTATE_AVX)))
-    {
-        trace("AVX not present\n");
-        feature_offsets[2] = 0;
-        feature_sizes[2] = 0;
-        xstate_part_size = offsetof( XSTATE, YmmContext );
-        supported_xstate_features = (1 << XSTATE_LEGACY_FLOATING_POINT) | (1 << XSTATE_LEGACY_SSE);
-    }
-
     trace("XState EnabledFeatures %#I64x, EnabledSupervisorFeatures %#I64x, EnabledVolatileFeatures %I64x.\n",
             xstate.EnabledFeatures, xstate.EnabledSupervisorFeatures, xstate.EnabledVolatileFeatures);
     feature_mask = pRtlGetEnabledExtendedFeatures(0);
@@ -2129,17 +2118,16 @@
     feature_mask = pGetEnabledXStateFeatures();
     ok(feature_mask == (xstate.EnabledFeatures | xstate.EnabledSupervisorFeatures), "Got unexpected feature_mask %s.\n",
             wine_dbgstr_longlong(feature_mask));
-    ok((xstate.EnabledFeatures & supported_xstate_features) == supported_xstate_features,
+    ok((xstate.EnabledFeatures & SUPPORTED_XSTATE_FEATURES) == SUPPORTED_XSTATE_FEATURES,
             "Got unexpected EnabledFeatures %s.\n", wine_dbgstr_longlong(xstate.EnabledFeatures));
-    ok((xstate.EnabledVolatileFeatures & supported_xstate_features) == (xstate.EnabledFeatures & supported_xstate_features),
+    ok((xstate.EnabledVolatileFeatures & SUPPORTED_XSTATE_FEATURES) == (xstate.EnabledFeatures & SUPPORTED_XSTATE_FEATURES),
             "Got unexpected EnabledVolatileFeatures %s.\n", wine_dbgstr_longlong(xstate.EnabledVolatileFeatures));
-    ok(xstate.Size >= 512 + xstate_part_size,
-            "Got unexpected Size %lu, expected %lu.\n", xstate.Size, (ULONG)(512 + xstate_part_size));
+    ok(xstate.Size >= 512 + sizeof(XSTATE), "Got unexpected Size %lu.\n", xstate.Size);
     if (xstate.CompactionEnabled)
         ok(xstate.OptimizedSave, "Got zero OptimizedSave with compaction enabled.\n");
     ok(!xstate.AlignedFeatures, "Got unexpected AlignedFeatures %s.\n",
             wine_dbgstr_longlong(xstate.AlignedFeatures));
-    ok(xstate.AllFeatureSize >= 512 + xstate_part_size
+    ok(xstate.AllFeatureSize >= 512 + sizeof(XSTATE)
             || !xstate.AllFeatureSize /* win8 on CPUs without XSAVEC */,
             "Got unexpected AllFeatureSize %lu.\n", xstate.AllFeatureSize);
 
@@ -2309,207 +2297,6 @@
     UnmapViewOfFile( ptr );
 }
 
-static void test_invalid_syscalls(void)
-{
-    HMODULE module = GetModuleHandleW( L"ntdll.dll" );
-    NTSTATUS (WINAPI *pNtImpersonateAnonymousToken)( HANDLE thread );
-    NTSTATUS status;
-    DWORD prot, i;
-    LONG old_id, new_id, *id;
-
-    /* grab a syscall that's unlikely to be used while we are testing */
-    pNtImpersonateAnonymousToken = (void *)GetProcAddress( module, "NtImpersonateAnonymousToken" );
-    if (!pNtImpersonateAnonymousToken)
-    {
-        win_skip( "NtImpersonateAnonymousToken not supported\n" );
-        return;
-    }
-    status = pNtImpersonateAnonymousToken( 0 );
-    ok( status == STATUS_INVALID_HANDLE || status == STATUS_NOT_IMPLEMENTED, "wrong status %lx\n", status );
-    VirtualProtect( pNtImpersonateAnonymousToken, 32, PAGE_EXECUTE_READWRITE, &prot );
-    for (i = 0; i < 4; i++)
-    {
-        new_id = 0x666 | (i << 12);
-        winetest_push_context( "%04lx", new_id );
-#ifdef __i386__
-        id = (LONG *)((BYTE *)pNtImpersonateAnonymousToken + 1);
-        new_id = (*id & ~0xffff) | new_id;
-#elif defined __x86_64__
-        id = (LONG *)pNtImpersonateAnonymousToken + 1;
-        new_id = (*id & ~0xffff) | new_id;
-#elif defined __aarch64__
-        id = (LONG *)pNtImpersonateAnonymousToken;
-        new_id = (*id & ~(0xffff << 5)) | (new_id << 5);
-#elif defined __arm__
-        id = (LONG *)(((ULONG_PTR)pNtImpersonateAnonymousToken & ~1) + 2);
-        new_id = 0x0c00f240 | ((new_id & 0xff) << 16) | ((new_id & 0xf00) << 20) | (new_id >> 12); /* movw ip, #0xnnn */
-#endif
-        old_id = *id;
-        *id = new_id;
-        NtFlushInstructionCache( GetCurrentProcess(), pNtImpersonateAnonymousToken, 32 );
-        status = pNtImpersonateAnonymousToken( 0 );
-        ok( status == STATUS_INVALID_SYSTEM_SERVICE, "wrong status %lx\n", status );
-        *id = old_id;
-        NtFlushInstructionCache( GetCurrentProcess(), pNtImpersonateAnonymousToken, 32 );
-        winetest_pop_context();
-    }
-    VirtualProtect( pNtImpersonateAnonymousToken, 32, prot, &prot );
-}
-
-struct syscall_export
-{
-    UINT        rva;
-    int         id;
-    const char *name;
-};
-
-static int CDECL sort_syscalls( const void *a, const void *b )
-{
-    const struct syscall_export *exp_a = a;
-    const struct syscall_export *exp_b = b;
-    int ret = exp_a->rva - exp_b->rva;
-    if (!ret) ret = strcmp( exp_a->name, exp_b->name );
-    return ret;
-}
-
-static int get_syscall_id( void *code )
-{
-#ifdef __i386__
-    static const BYTE patterns[][18] =
-    {
-        { 0xb8, 0, 0, 0, 0, 0xba, 0, 0, 0, 0, 0xff, 0xd2 }, /* >= win10 */
-        { 0xb8, 0, 0, 0, 0, 0xba, 0, 0, 0, 0, 0xff, 0x12 }, /* winxp */
-        { 0xb8, 0, 0, 0, 0, 0x64, 0xff, 0x15, 0xc0, 0, 0, 0 },  /* nt */
-        { 0xb8, 0, 0, 0, 0, 0x8d, 0x54, 0x24, 0x04, 0xcd, 0x2e }, /* nt */
-        { 0xb8, 0, 0, 0, 0, 0xb9, 0, 0, 0, 0, 0x8d, 0x54, 0x24, 0x04, 0x64, 0xff, 0x15, 0xc0 }, /* vista */
-        { 0xb8, 0, 0, 0, 0, 0x33, 0xc9, 0x8d, 0x54, 0x24, 0x04, 0x64, 0xff, 0x15, 0xc0 }, /* vista */
-        { 0xb8, 0, 0, 0, 0, 0xe8, 0, 0, 0, 0, 0x8d, 0x54, 0x24, 0x04, 0x64, 0xff, 0x15, 0xc0 }, /* win8 */
-        { 0xb8, 0, 0, 0, 0, 0xe8, 0x01, 0, 0, 0, 0xc3, 0x8b, 0xd4, 0x0f, 0x34, 0xc3 },  /* win8 */
-        { 0xb8, 0, 0, 0, 0, 0xe8, 0x03, 0, 0, 0, 0xc2, 0, 0, 0x8b, 0xd4, 0x0f, 0x34, 0xc3 }, /* win8 */
-    };
-    const BYTE *instr = code;
-    UINT i, j;
-
-    for (i = 0; i < ARRAY_SIZE(patterns); i++)
-    {
-        for (j = 0; j < ARRAY_SIZE(patterns[0]); j++)
-            if (patterns[i][j] && patterns[i][j] != instr[j]) break;
-        if (j == ARRAY_SIZE(patterns[0]))
-            return *(UINT *)(instr + 1);
-    }
-#elif defined __x86_64__
-    static const BYTE patterns[][20] =
-    {
-        { 0x4c, 0x8b, 0xd1, 0xb8, 0, 0, 0, 0, 0xf6, 0x04, 0x25, 0x08, 0x03, 0xfe,
-          0x7f, 0x01, 0x75, 0x03, 0x0f, 0x05 },  /* >= win10 */
-        { 0x4c, 0x8b, 0xd1, 0xb8, 0, 0, 0, 0, 0x0f, 0x05, 0xc3 }, /* < win10 */
-    };
-    const BYTE *instr = code;
-    UINT i, j;
-
-    for (i = 0; i < ARRAY_SIZE(patterns); i++)
-    {
-        for (j = 0; j < ARRAY_SIZE(patterns[0]); j++)
-            if (patterns[i][j] && patterns[i][j] != instr[j]) break;
-        if (j == ARRAY_SIZE(patterns[0]))
-            return ((UINT *)instr)[1];
-    }
-#elif defined __aarch64__
-    const UINT *instr = code;
-
-    if ((instr[0] & 0xffe0001f) == 0xd4000001 && instr[1] == 0xd65f03c0)  /* windows */
-        return (instr[0] >> 5) & 0xffff;
-    if ((instr[0] & 0xffe0001f) == 0xd2800008 && instr[1] == 0xaa1e03e9 &&
-        instr[3] == 0xf9400210 && instr[4] == 0xd63f0200 && instr[5] == 0xd65f03c0) /* wine */
-        return (instr[0] >> 5) & 0xffff;
-#elif defined __arm__
-    const USHORT *instr = code;
-
-    if (instr[0] == 0xb40f && (instr[2] & 0x0f00) == 0x0c00 &&
-        ((instr[3] == 0xdef8 && instr[4] == 0xb004 && instr[5] == 0x4770) ||  /* windows */
-         (instr[3] == 0x4673 && instr[6] == 0xb004 && instr[7] == 0x4770)))  /* wine */
-    {
-        USHORT imm = ((instr[1] & 0x400) << 1) | (instr[2] & 0xff) | ((instr[2] >> 4) & 0x0700);
-        if ((instr[1] & 0xfbf0) == 0xf240)  /* T3 */
-        {
-            return imm | (instr[1] & 0x0f) << 12;
-        }
-        else if ((instr[1] & 0xfbf0) == 0xf040)  /* T2 */
-        {
-            switch (imm >> 8)
-            {
-            case 0: return imm;
-            case 1: return (imm & 0xff);
-            case 2: return (imm & 0xff) << 8;
-            case 3: return (imm & 0xff) | ((imm & 0xff) << 8);
-            default: return (0x80 | (imm & 0x7f)) << (32 - (imm >> 7));
-            }
-        }
-    }
-#endif
-    return -1;
-}
-
-static void test_syscall_numbers(void)
-{
-    struct syscall_export syscalls[4096];
-    HMODULE module = GetModuleHandleA( "ntdll.dll" );
-    IMAGE_EXPORT_DIRECTORY *exports;
-    ULONG size;
-    int pos;
-    const WORD *ordinals;
-    const DWORD *names, *functions;
-    static const char *prefix[] = { "Nt", "Zw" };
-
-    exports = RtlImageDirectoryEntryToData( module, TRUE, IMAGE_DIRECTORY_ENTRY_EXPORT, &size );
-    names = get_rva( module, exports->AddressOfNames );
-    ordinals = get_rva( module, exports->AddressOfNameOrdinals );
-    functions = get_rva( module, exports->AddressOfFunctions );
-
-    for (unsigned int test = 0; test < ARRAY_SIZE(prefix); test++)
-    {
-        for (int i = pos = 0; i < exports->NumberOfNames; i++)
-        {
-            char *name = get_rva( module, names[i] );
-            if (strncmp( name, prefix[test], strlen(prefix[test]) )) continue;
-            if (!strcmp( name, "NtGetTickCount" ) ||
-                !strcmp( name, "NtCurrentTeb" ) ||
-                !strncmp( name, "NtdllDialogWndProc", 18 ) ||
-                !strncmp( name, "NtdllDefWindowProc", 18 ))
-                continue;  /* these are special */
-            syscalls[pos].rva  = functions[ordinals[i]];
-#ifdef __arm__
-            syscalls[pos].rva &= ~1; /* thumb */
-#endif
-            syscalls[pos].id   = get_syscall_id( get_rva( module, syscalls[pos].rva ));
-            syscalls[pos].name = name;
-            pos++;
-        }
-        ok( pos, "no syscalls found\n" );
-        qsort( syscalls, pos, sizeof(*syscalls), sort_syscalls );
-        for (int i = 0, expect = 0; i < pos; i++, expect++)
-        {
-            if (syscalls[i].id == -1)
-            {
-                /* these may not be real syscalls */
-                ok( !strcmp( syscalls[i].name + strlen(prefix[test]), "QuerySystemTime" ) ||
-                    !strcmp( syscalls[i].name + strlen(prefix[test]), "QueryInformationProcess" ),
-                    "not a syscall %04x %s\n", i, syscalls[i].name );
-            }
-            else if (LOWORD(syscalls[i].id) > expect)
-            {
-                ok( 0, "missing syscall %04x\n", expect );
-                i--;
-            }
-            else
-            {
-                ok( LOWORD(syscalls[i].id) == expect, "wrong id %04x / %04x for %s\n",
-                    syscalls[i].id, expect, syscalls[i].name );
-            }
-        }
-    }
-}
-
 static void test_NtFreeVirtualMemory(void)
 {
     void *addr1, *addr;
@@ -2590,80 +2377,6 @@
 
     size = 0x1000;
     status = NtFreeVirtualMemory(NtCurrentProcess(), &addr1, &size, MEM_RELEASE);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-}
-
-static void test_NtProtectVirtualMemory(void)
-{
-    void *addr, *addr2;
-    NTSTATUS status;
-    SIZE_T size;
-    DWORD old_prot;
-
-    size = page_size * 16;
-    addr = NULL;
-    status = NtAllocateVirtualMemory(NtCurrentProcess(), &addr, 0, &size, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-
-    old_prot = 0;
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_READONLY, &old_prot);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(old_prot == PAGE_READWRITE, "Unexpected old_prot %lx.\n", old_prot);
-
-    status = NtProtectVirtualMemory(NULL, &addr, &size, PAGE_READONLY, &old_prot);
-    ok(status == STATUS_INVALID_HANDLE, "Unexpected status %08lx.\n", status);
-
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_READONLY, NULL);
-    ok(status == STATUS_ACCESS_VIOLATION, "Unexpected status %08lx.\n", status);
-
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, 0, &old_prot);
-    ok(status == STATUS_INVALID_PAGE_PROTECTION, "Unexpected status %08lx.\n", status);
-
-    size = page_size * 8;
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_READWRITE, &old_prot);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(old_prot == PAGE_READONLY, "Unexpected old_prot %lx.\n", old_prot);
-    addr = (char *)addr + page_size * 8;
-    old_prot = 0;
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_EXECUTE_READWRITE, &old_prot);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(old_prot == PAGE_READONLY, "Unexpected old_prot %lx.\n", old_prot);
-    addr = (char *)addr - page_size * 8;
-    size = page_size * 16;
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_EXECUTE_READ, &old_prot);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(old_prot == PAGE_READWRITE, "Unexpected old_prot %lx.\n", old_prot);
-
-    status = NtFreeVirtualMemory(NtCurrentProcess(), &addr, &size, MEM_RELEASE);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-
-    addr = NULL;
-    size = page_size;
-    old_prot = 0;
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_READONLY, &old_prot);
-    /* todo: STATUS_INVALID_PARAMETER in wine, STATUS_CONFLICTING_ADDRESSES in win64 */
-    ok(status, "Unexpected status %08lx.\n", status);
-    ok(old_prot == PAGE_NOACCESS || broken(old_prot) /* win7 */, "Unexpected old_prot %lx.\n", old_prot);
-
-    status = NtAllocateVirtualMemory(NtCurrentProcess(), &addr, 0, &size, MEM_RESERVE, PAGE_READWRITE);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    old_prot = 0;
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_READONLY, &old_prot);
-    ok(status == STATUS_NOT_COMMITTED, "Unexpected status %08lx.\n", status);
-    ok(old_prot == PAGE_NOACCESS || broken(old_prot) /* win7 */, "Unexpected old_prot %lx.\n", old_prot);
-    status = NtAllocateVirtualMemory(NtCurrentProcess(), &addr, 0, &size, MEM_COMMIT, PAGE_READWRITE);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-
-    addr2 = addr;
-    addr = (char *)addr + 1;
-    size = 1;
-    status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_READONLY, &old_prot);
-    ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(old_prot == PAGE_READWRITE, "Unexpected old_prot %lx.\n", old_prot);
-    ok(size == page_size, "Unexpected size %p.\n",  (void *)size);
-    ok(addr == addr2, "Got addr %p, addr2 %p.\n", addr, addr2);
-
-    status = NtFreeVirtualMemory(NtCurrentProcess(), &addr, &size, MEM_RELEASE);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
 }
 
@@ -2821,10 +2534,6 @@
     memset(&info, 0x11, sizeof(info));
     status = NtQueryVirtualMemory(NtCurrentProcess(), ptr, MemoryRegionInformation, &info, sizeof(info), &len);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(len == sizeof(info) ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId)) /* <= Win10-1909 */ ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, CommitSize)) /* Win7 */,
-       "Unexpected len %Ix\n", len);
     ok(info.AllocationBase == ptr, "Unexpected base %p.\n", info.AllocationBase);
     ok(info.AllocationProtect == PAGE_READWRITE, "Unexpected protection %lu.\n", info.AllocationProtect);
     ok(!info.Private, "Unexpected flag %d.\n", info.Private);
@@ -2834,8 +2543,7 @@
     ok(!info.MappedPhysical, "Unexpected flag %d.\n", info.MappedPhysical);
     ok(!info.DirectMapped, "Unexpected flag %d.\n", info.DirectMapped);
     ok(info.RegionSize == size, "Unexpected region size.\n");
-    if (len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId))
-        ok(!info.CommitSize, "Unexpected commit size %#Ix.\n", info.CommitSize);
+    ok(!info.CommitSize, "Unexpected commit size %#Ix.\n", info.CommitSize);
 
     size = 0;
     status = NtFreeVirtualMemory(NtCurrentProcess(), &ptr, &size, MEM_RELEASE);
@@ -2847,14 +2555,9 @@
     status = NtAllocateVirtualMemory(NtCurrentProcess(), &ptr, 0, &size, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
 
-    len = 0;
     memset(&info, 0x11, sizeof(info));
     status = NtQueryVirtualMemory(NtCurrentProcess(), ptr, MemoryRegionInformation, &info, sizeof(info), &len);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(len == sizeof(info) ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId)) /* <= Win10-1909 */ ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, CommitSize)) /* Win7 */,
-       "Unexpected len %Ix\n", len);
     ok(info.AllocationBase == ptr, "Unexpected base %p.\n", info.AllocationBase);
     ok(info.AllocationProtect == PAGE_READWRITE, "Unexpected protection %lu.\n", info.AllocationProtect);
     ok(!info.Private, "Unexpected flag %d.\n", info.Private);
@@ -2864,20 +2567,14 @@
     ok(!info.MappedPhysical, "Unexpected flag %d.\n", info.MappedPhysical);
     ok(!info.DirectMapped, "Unexpected flag %d.\n", info.DirectMapped);
     ok(info.RegionSize == size, "Unexpected region size.\n");
-    if (len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId))
-        ok(info.CommitSize == size, "Unexpected commit size %#Ix.\n", info.CommitSize);
-
-    len = 0;
+    ok(info.CommitSize == size, "Unexpected commit size %#Ix.\n", info.CommitSize);
+
     addr = (char *)ptr + 0x1000;
     size = 0x1000;
     status = NtProtectVirtualMemory(NtCurrentProcess(), &addr, &size, PAGE_NOACCESS, &old );
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
     status = NtQueryVirtualMemory(NtCurrentProcess(), ptr, MemoryRegionInformation, &info, sizeof(info), &len);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(len == sizeof(info) ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId)) /* <= Win10-1909 */ ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, CommitSize)) /* Win7 */,
-       "Unexpected len %Ix\n", len);
     ok(info.AllocationBase == ptr, "Unexpected base %p.\n", info.AllocationBase);
     ok(info.AllocationProtect == PAGE_READWRITE, "Unexpected protection %lu.\n", info.AllocationProtect);
     ok(!info.Private, "Unexpected flag %d.\n", info.Private);
@@ -2887,16 +2584,10 @@
     ok(!info.MappedPhysical, "Unexpected flag %d.\n", info.MappedPhysical);
     ok(!info.DirectMapped, "Unexpected flag %d.\n", info.DirectMapped);
     ok(info.RegionSize == 0x10000, "Unexpected region size %#Ix.\n", info.RegionSize);
-    if (len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId))
-        ok(info.CommitSize == 0x10000, "Unexpected commit size %#Ix.\n", info.CommitSize);
-
-    len = 0;
+    ok(info.CommitSize == 0x10000, "Unexpected commit size %#Ix.\n", info.CommitSize);
+
     status = NtQueryVirtualMemory(NtCurrentProcess(), (char *)ptr + 0x1000, MemoryRegionInformation, &info, sizeof(info), &len);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(len == sizeof(info) ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId)) /* <= Win10-1909 */ ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, CommitSize)) /* Win7 */,
-       "Unexpected len %Ix\n", len);
     ok(info.AllocationBase == ptr, "Unexpected base %p.\n", info.AllocationBase);
     ok(info.AllocationProtect == PAGE_READWRITE, "Unexpected protection %lu.\n", info.AllocationProtect);
     ok(!info.Private, "Unexpected flag %d.\n", info.Private);
@@ -2906,8 +2597,7 @@
     ok(!info.MappedPhysical, "Unexpected flag %d.\n", info.MappedPhysical);
     ok(!info.DirectMapped, "Unexpected flag %d.\n", info.DirectMapped);
     ok(info.RegionSize == 0x10000, "Unexpected region size %#Ix.\n", info.RegionSize);
-    if (len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId))
-        ok(info.CommitSize == 0x10000, "Unexpected commit size %#Ix.\n", info.CommitSize);
+    ok(info.CommitSize == 0x10000, "Unexpected commit size %#Ix.\n", info.CommitSize);
 
     size = 0;
     status = NtFreeVirtualMemory(NtCurrentProcess(), &ptr, &size, MEM_RELEASE);
@@ -2930,14 +2620,9 @@
     status = NtMapViewOfSection(mapping, NtCurrentProcess(), &ptr, 0, 0, &offset, &size, 1, 0, PAGE_READONLY);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
 
-    len = 0;
     memset(&info, 0x11, sizeof(info));
     status = NtQueryVirtualMemory(NtCurrentProcess(), ptr, MemoryRegionInformation, &info, sizeof(info), &len);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(len == sizeof(info) ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId)) /* <= Win10-1909 */ ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, CommitSize)) /* Win7 */,
-       "Unexpected len %Ix\n", len);
     ok(info.AllocationBase == ptr, "Unexpected base %p.\n", info.AllocationBase);
     ok(info.AllocationProtect == PAGE_READONLY, "Unexpected protection %lu.\n", info.AllocationProtect);
     ok(!info.Private, "Unexpected flag %d.\n", info.Private);
@@ -2947,8 +2632,7 @@
     ok(!info.MappedPhysical, "Unexpected flag %d.\n", info.MappedPhysical);
     ok(!info.DirectMapped, "Unexpected flag %d.\n", info.DirectMapped);
     ok(info.RegionSize == 4096, "Unexpected region size.\n");
-    if (len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId))
-        ok(!info.CommitSize, "Unexpected commit size %#Ix.\n", info.CommitSize);
+    ok(!info.CommitSize, "Unexpected commit size %#Ix.\n", info.CommitSize);
 
     status = NtUnmapViewOfSection(NtCurrentProcess(), ptr);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
@@ -2959,14 +2643,9 @@
     status = NtMapViewOfSection(mapping, NtCurrentProcess(), &ptr, 0, 0, &offset, &size, 1, 0, PAGE_WRITECOPY);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
 
-    len = 0;
     memset(&info, 0x11, sizeof(info));
     status = NtQueryVirtualMemory(NtCurrentProcess(), ptr, MemoryRegionInformation, &info, sizeof(info), &len);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(len == sizeof(info) ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId)) /* <= Win10-1909 */ ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, CommitSize)) /* Win7 */,
-       "Unexpected len %Ix\n", len);
     ok(info.AllocationBase == ptr, "Unexpected base %p.\n", info.AllocationBase);
     ok(info.AllocationProtect == PAGE_WRITECOPY, "Unexpected protection %lu.\n", info.AllocationProtect);
     ok(!info.Private, "Unexpected flag %d.\n", info.Private);
@@ -2976,18 +2655,12 @@
     ok(!info.MappedPhysical, "Unexpected flag %d.\n", info.MappedPhysical);
     ok(!info.DirectMapped, "Unexpected flag %d.\n", info.DirectMapped);
     ok(info.RegionSize == 4096, "Unexpected region size.\n");
-    if (len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId))
-        ok(info.CommitSize == 4096, "Unexpected commit size %#Ix.\n", info.CommitSize);
-
-    len = 0;
+    ok(info.CommitSize == 4096, "Unexpected commit size %#Ix.\n", info.CommitSize);
+
     *(volatile int *)ptr = 1;
     memset(&info, 0x11, sizeof(info));
     status = NtQueryVirtualMemory(NtCurrentProcess(), ptr, MemoryRegionInformation, &info, sizeof(info), &len);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(len == sizeof(info) ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId)) /* <= Win10-1909 */ ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, CommitSize)) /* Win7 */,
-       "Unexpected len %Ix\n", len);
     ok(info.AllocationBase == ptr, "Unexpected base %p.\n", info.AllocationBase);
     ok(info.AllocationProtect == PAGE_WRITECOPY, "Unexpected protection %lu.\n", info.AllocationProtect);
     ok(!info.Private, "Unexpected flag %d.\n", info.Private);
@@ -2997,8 +2670,7 @@
     ok(!info.MappedPhysical, "Unexpected flag %d.\n", info.MappedPhysical);
     ok(!info.DirectMapped, "Unexpected flag %d.\n", info.DirectMapped);
     ok(info.RegionSize == 4096, "Unexpected region size.\n");
-    if (len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId))
-        ok(info.CommitSize == 4096, "Unexpected commit size %#Ix.\n", info.CommitSize);
+    ok(info.CommitSize == 4096, "Unexpected commit size %#Ix.\n", info.CommitSize);
 
     status = NtUnmapViewOfSection(NtCurrentProcess(), ptr);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
@@ -3010,14 +2682,9 @@
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
     *(volatile int *)ptr = 1;
 
-    len = 0;
     memset(&info, 0x11, sizeof(info));
     status = NtQueryVirtualMemory(NtCurrentProcess(), ptr, MemoryRegionInformation, &info, sizeof(info), &len);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
-    ok(len == sizeof(info) ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId)) /* <= Win10-1909 */ ||
-       broken(len >= offsetof(MEMORY_REGION_INFORMATION, CommitSize)) /* Win7 */,
-       "Unexpected len %Ix\n", len);
     ok(info.AllocationBase == ptr, "Unexpected base %p.\n", info.AllocationBase);
     ok(info.AllocationProtect == PAGE_READWRITE, "Unexpected protection %lu.\n", info.AllocationProtect);
     ok(!info.Private, "Unexpected flag %d.\n", info.Private);
@@ -3027,8 +2694,7 @@
     ok(!info.MappedPhysical, "Unexpected flag %d.\n", info.MappedPhysical);
     ok(!info.DirectMapped, "Unexpected flag %d.\n", info.DirectMapped);
     ok(info.RegionSize == 4096, "Unexpected region size.\n");
-    if (len >= offsetof(MEMORY_REGION_INFORMATION, PartitionId))
-        ok(!info.CommitSize, "Unexpected commit size %#Ix.\n", info.CommitSize);
+    ok(!info.CommitSize, "Unexpected commit size %#Ix.\n", info.CommitSize);
 
     status = NtUnmapViewOfSection(NtCurrentProcess(), ptr);
     ok(status == STATUS_SUCCESS, "Unexpected status %08lx.\n", status);
@@ -3565,15 +3231,12 @@
     test_NtAllocateVirtualMemoryEx();
     test_NtAllocateVirtualMemoryEx_address_requirements();
     test_NtFreeVirtualMemory();
-    test_NtProtectVirtualMemory();
     test_RtlCreateUserStack();
     test_NtMapViewOfSection();
     test_NtMapViewOfSectionEx();
     test_prefetch();
     test_user_shared_data();
     test_syscalls();
-    test_invalid_syscalls();
-    test_syscall_numbers();
     test_query_region_information();
     test_query_image_information();
     test_exec_memory_writes();
