diff --git a/dlls/sane.ds/ds_image.c b/dlls/sane.ds/ds_image.c
--- a/dlls/sane.ds/ds_image.c
+++ b/dlls/sane.ds/ds_image.c
@@ -23,193 +23,8 @@
 #include "wingdi.h"
 #include "winuser.h"
 #include "wine/debug.h"
-#include "resource.h"
-
-#include <winnls.h>
 
 WINE_DEFAULT_DEBUG_CHANNEL(twain);
-
-/* Sane result codes from sane.h */
-typedef enum
-{
-    SANE_STATUS_GOOD = 0,       /* everything A-OK */
-    SANE_STATUS_UNSUPPORTED,    /* operation is not supported */
-    SANE_STATUS_CANCELLED,      /* operation was cancelled */
-    SANE_STATUS_DEVICE_BUSY,    /* device is busy; try again later */
-    SANE_STATUS_INVAL,          /* data is invalid (includes no dev at open) */
-    SANE_STATUS_EOF,            /* no more data available (end-of-file) */
-    SANE_STATUS_JAMMED,         /* document feeder jammed */
-    SANE_STATUS_NO_DOCS,        /* document feeder out of documents */
-    SANE_STATUS_COVER_OPEN,     /* scanner cover is open */
-    SANE_STATUS_IO_ERROR,       /* error during device I/O */
-    SANE_STATUS_NO_MEM,         /* out of memory */
-    SANE_STATUS_ACCESS_DENIED   /* access to resource has been denied */
-}
-SANE_Status;
-
-
-/* transition from state 6 to state 7.
- *
- * Called by either SANE_ImageMemXferGet, SANE_ImageInfoGet or
- * SANE_ImageNativeXferGet, whatever the application calls first.
- *
- * - start the scan with a call to sane start_devince
- * - open the progress dialog window
- * - call get_sane_params to retrieve parameters of current scan frame
- *
- * @return TWAIN result code, TWRC_SUCCESS on success
- */
-TW_UINT16 SANE_Start(void)
-{
-  TW_UINT16 twRC = TWRC_SUCCESS;
-  SANE_Status status;
-
-  TRACE("SANE_Start currentState:%d\n", activeDS.currentState);
-  if (activeDS.currentState != 6)
-  {
-      twRC = TWRC_FAILURE;
-      activeDS.twCC = TWCC_SEQERROR;
-  }
-  else
-  {
-      /* Open progress dialog */
-      activeDS.progressWnd = ScanningDialogBox(activeDS.progressWnd,0);
-
-      do
-      {
-          /* Start the scan process in sane */
-          status = SANE_CALL( start_device, NULL );
-
-          if (status == SANE_STATUS_GOOD)
-          {
-              twRC = TWRC_SUCCESS;
-          }
-          else if (!activeDS.capIndicators && !activeDS.ShowUI)
-          {
-              twRC = TWRC_FAILURE;
-          }
-          else
-          {
-              WCHAR szCaption[256];
-              WCHAR szMessage[1024];
-
-              LoadStringW( SANE_instance, IDS_CAPTION, szCaption, ARRAY_SIZE( szCaption ) );
-              switch (status)
-              {
-                case SANE_STATUS_NO_DOCS:
-                  LoadStringW( SANE_instance, IDS_NO_DOCS, szMessage, ARRAY_SIZE( szMessage ) );
-                  twRC =
-                    activeDS.scannedImages!=0 ? TWRC_FAILURE :
-                    MessageBoxW(activeDS.progressWnd, szMessage, szCaption, MB_ICONWARNING | MB_RETRYCANCEL)==IDCANCEL ? TWRC_FAILURE :TWRC_CHECKSTATUS;
-                  break;
-                case SANE_STATUS_JAMMED:
-                  LoadStringW( SANE_instance, IDS_JAMMED, szMessage, ARRAY_SIZE( szMessage ) );
-                  twRC =
-                    MessageBoxW(activeDS.progressWnd, szMessage, szCaption, MB_ICONWARNING | MB_RETRYCANCEL)==IDCANCEL ? TWRC_FAILURE :TWRC_CHECKSTATUS;
-                  break;
-                case SANE_STATUS_COVER_OPEN:
-                  LoadStringW( SANE_instance, IDS_COVER_OPEN, szMessage, ARRAY_SIZE( szMessage ) );
-                  twRC =
-                    MessageBoxW(activeDS.progressWnd, szMessage, szCaption, MB_ICONWARNING | MB_RETRYCANCEL)==IDCANCEL ? TWRC_FAILURE :TWRC_CHECKSTATUS;
-                  break;
-                case SANE_STATUS_DEVICE_BUSY:
-                  LoadStringW( SANE_instance, IDS_DEVICE_BUSY, szMessage, ARRAY_SIZE( szMessage ) );
-                  twRC =
-                    MessageBoxW(activeDS.progressWnd, szMessage, szCaption, MB_ICONWARNING | MB_RETRYCANCEL)==IDCANCEL ? TWRC_FAILURE :TWRC_CHECKSTATUS;
-                  break;
-                default:
-                  twRC = TWRC_FAILURE;
-              }
-          }
-      }
-      while (twRC == TWRC_CHECKSTATUS);
-
-      /* If starting the scan failed, cancel scan job */
-      if (twRC != TWRC_SUCCESS)
-      {
-          UI_Enable(TRUE);
-          activeDS.progressWnd = ScanningDialogBox(activeDS.progressWnd, -1);
-          activeDS.twCC = TWCC_OPERATIONERROR;
-          return TWRC_FAILURE;
-      }
-
-      if (get_sane_params( &activeDS.frame_params ))
-      {
-          WARN("sane_get_parameters failed\n");
-          SANE_CALL( cancel_device, NULL );
-          UI_Enable(TRUE);
-          activeDS.progressWnd = ScanningDialogBox(activeDS.progressWnd, -1);
-          activeDS.twCC = TWCC_OPERATIONERROR;
-          return TWRC_FAILURE;
-      }
-
-      if (activeDS.progressWnd)
-      {
-          WCHAR szLocaleBuffer[4];
-          WCHAR szResolution[200];
-          WCHAR szFormat[20];
-          int sid_format;
-
-          switch (activeDS.frame_params.format)
-          {
-          case FMT_RGB:  sid_format=IDS_COLOUR; break;
-          case FMT_GRAY: sid_format=activeDS.frame_params.depth==1 ? IDS_LINEART : IDS_GRAY; break;
-          default:       sid_format=IDS_UNKNOWN;
-          }
-
-          LoadStringW( SANE_instance, sid_format, szFormat, ARRAY_SIZE( szFormat ) );
-
-          if (activeDS.XResolution.Whole && activeDS.YResolution.Whole)
-          {
-              const double xresolution = activeDS.XResolution.Whole + activeDS.XResolution.Frac / 65536.0;
-              const double yresolution = activeDS.YResolution.Whole + activeDS.YResolution.Frac / 65536.0;
-
-              GetLocaleInfoEx(LOCALE_NAME_USER_DEFAULT, LOCALE_IMEASURE, szLocaleBuffer, ARRAY_SIZE(szLocaleBuffer));
-              if (szLocaleBuffer[0]=='1')
-              {
-                  swprintf(szResolution, ARRAY_SIZE(szResolution), L"%d DPI %s %.2f x %.2f \"",
-                           activeDS.XResolution.Whole, szFormat,
-                           activeDS.frame_params.pixels_per_line / xresolution ,
-                           activeDS.frame_params.lines           / yresolution );
-              }
-              else
-              {
-                  swprintf(szResolution, ARRAY_SIZE(szResolution), L"%d DPI %s %.1f x %.1f mm",
-                           activeDS.XResolution.Whole, szFormat,
-                           activeDS.frame_params.pixels_per_line * 25.4 / xresolution ,
-                           activeDS.frame_params.lines           * 25.4 / yresolution );
-              }
-              SetDlgItemTextW(activeDS.progressWnd, IDC_RESOLUTION, szResolution);
-          }
-      }
-
-      TRACE("Acquiring image %dx%dx%d bits (format=%d last=%d) from sane...\n"
-            , activeDS.frame_params.pixels_per_line, activeDS.frame_params.lines,
-            activeDS.frame_params.depth, activeDS.frame_params.format,
-            activeDS.frame_params.last_frame);
-
-      activeDS.currentState = 7;
-      activeDS.YOffset = 0;
-  }
-  return twRC;
-}
-
-/** transition from state 7 or 6 to state 5.
- *
- * Called when an error occurs or when the last frame
- * has been transfered successfully.
- * - call sane cancel_device to finish any open transfer
- * - close the progress dialog box.
- */
-void SANE_Cancel(void)
-{
-  SANE_CALL( cancel_device, NULL );
-  UI_Enable(TRUE);
-  activeDS.progressWnd = ScanningDialogBox(activeDS.progressWnd, -1);
-  activeDS.currentState = 5;
-}
-
-
 
 /* DG_IMAGE/DAT_IMAGEINFO/MSG_GET */
 TW_UINT16 SANE_ImageInfoGet (pTW_IDENTITY pOrigin, 
@@ -217,6 +32,7 @@
 {
     TW_UINT16 twRC = TWRC_SUCCESS;
     pTW_IMAGEINFO pImageInfo = (pTW_IMAGEINFO) pData;
+    int resolution;
 
     TRACE("DG_IMAGE/DAT_IMAGEINFO/MSG_GET\n");
 
@@ -229,16 +45,23 @@
     {
         if (activeDS.currentState == 6)
         {
-            /* Transition from state 6 to state 7 by starting the scan */
-            twRC = SANE_Start();
-            if (twRC != TWRC_SUCCESS)
-            {
-                return twRC;
-            }
-        }
-
-        pImageInfo->XResolution = activeDS.XResolution;
-        pImageInfo->YResolution = activeDS.YResolution;
+            /* return general image description information about the image about to be transferred */
+            TRACE("Getting parameters\n");
+            if (SANE_CALL( get_params, &activeDS.frame_params ))
+            {
+                WARN("sane_get_parameters failed\n");
+                SANE_CALL( cancel_device, NULL );
+                activeDS.twCC = TWCC_OPERATIONERROR;
+                return TWRC_FAILURE;
+            }
+        }
+
+        if (sane_option_get_int("resolution", &resolution) == TWCC_SUCCESS)
+            pImageInfo->XResolution.Whole = pImageInfo->YResolution.Whole = resolution;
+        else
+            pImageInfo->XResolution.Whole = pImageInfo->YResolution.Whole = -1;
+        pImageInfo->XResolution.Frac = 0;
+        pImageInfo->YResolution.Frac = 0;
         pImageInfo->ImageWidth = activeDS.frame_params.pixels_per_line;
         pImageInfo->ImageLength = activeDS.frame_params.lines;
 
@@ -378,24 +201,47 @@
         /* Transfer an image from the source to the application */
         if (activeDS.currentState == 6)
         {
-            /* Transition from state 6 to state 7 by starting the scan */
-            twRC = SANE_Start();
-            if (twRC != TWRC_SUCCESS)
-            {
-                return twRC;
-            }
+
+            /* trigger scanning dialog */
+            activeDS.progressWnd = ScanningDialogBox(NULL,0);
+
+            ScanningDialogBox(activeDS.progressWnd,0);
+
+            if (SANE_CALL( start_device, NULL ))
+            {
+                activeDS.twCC = TWCC_OPERATIONERROR;
+                return TWRC_FAILURE;
+            }
+
+            if (get_sane_params( &activeDS.frame_params ))
+            {
+                WARN("sane_get_parameters failed\n");
+                SANE_CALL( cancel_device, NULL );
+                activeDS.twCC = TWCC_OPERATIONERROR;
+                return TWRC_FAILURE;
+            }
+
+            TRACE("Acquiring image %dx%dx%d bits (format=%d last=%d) from sane...\n"
+              , activeDS.frame_params.pixels_per_line, activeDS.frame_params.lines,
+              activeDS.frame_params.depth, activeDS.frame_params.format,
+              activeDS.frame_params.last_frame);
+
+            activeDS.currentState = 7;
         }
 
         /* access memory buffer */
         if (pImageMemXfer->Memory.Length < activeDS.frame_params.bytes_per_line)
         {
-            SANE_Cancel();
+            SANE_CALL( cancel_device, NULL );
             activeDS.twCC = TWCC_BADVALUE;
             return TWRC_FAILURE;
         }
 
         if (pImageMemXfer->Memory.Flags & TWMF_HANDLE)
-            buffer = GlobalLock(pImageMemXfer->Memory.TheMem);
+        {
+            FIXME("Memory Handle, may not be locked correctly\n");
+            buffer = LocalLock(pImageMemXfer->Memory.TheMem);
+        }
         else
             buffer = pImageMemXfer->Memory.TheMem;
        
@@ -415,37 +261,32 @@
             pImageMemXfer->Columns = activeDS.frame_params.pixels_per_line;
             pImageMemXfer->Rows = rows;
             pImageMemXfer->XOffset = 0;
-            pImageMemXfer->YOffset = activeDS.YOffset;
+            pImageMemXfer->YOffset = 0;
             pImageMemXfer->BytesWritten = retlen;
-            activeDS.YOffset += rows;
-
-            ScanningDialogBox(activeDS.progressWnd, MulDiv(activeDS.YOffset, 100, activeDS.frame_params.lines));
+
+            ScanningDialogBox(activeDS.progressWnd, retlen);
 
             if (retlen < activeDS.frame_params.bytes_per_line * rows)
             {
+                ScanningDialogBox(activeDS.progressWnd, -1);
                 TRACE("sane_read: %u / %u\n", retlen, activeDS.frame_params.bytes_per_line * rows);
+                SANE_CALL( cancel_device, NULL );
                 twRC = TWRC_XFERDONE;
             }
             activeDS.twCC = TWRC_SUCCESS;
         }
         else
         {
+            ScanningDialogBox(activeDS.progressWnd, -1);
             WARN("sane_read: %u\n", twRC);
-            SANE_Cancel();
+            SANE_CALL( cancel_device, NULL );
             activeDS.twCC = TWCC_OPERATIONERROR;
             twRC = TWRC_FAILURE;
         }
     }
 
     if (pImageMemXfer->Memory.Flags & TWMF_HANDLE)
-        GlobalUnlock(pImageMemXfer->Memory.TheMem);
-
-    if (activeDS.userCancelled)
-    {
-        SANE_Cancel();
-        activeDS.twCC = TWCC_OPERATIONERROR;
-        twRC = TWRC_FAILURE;
-    }
+        LocalUnlock(pImageMemXfer->Memory.TheMem);
     
     return twRC;
 }
@@ -455,7 +296,7 @@
                                     TW_MEMREF pData)
 {
     TW_UINT16 twRC = TWRC_SUCCESS;
-    TW_HANDLE *pHandle = (TW_HANDLE *) pData;
+    pTW_UINT32 pHandle = (pTW_UINT32) pData;
     HANDLE hDIB;
     BITMAPINFOHEADER *header = NULL;
     int dib_bytes;
@@ -465,29 +306,32 @@
     RGBTRIPLE *pixels;
     int color_size = 0;
     int i, j;
-    int y, eof = 0;
     BYTE *p;
-    DWORD tmp, *ptop, *pbot;
 
     TRACE("DG_IMAGE/DAT_IMAGENATIVEXFER/MSG_GET\n");
 
-    if (activeDS.currentState == 6)
-    {
-        /* Transition from state 6 to state 7 by starting the scan */
-        twRC = SANE_Start();
-        if (twRC != TWRC_SUCCESS)
-        {
-            return twRC;
-        }
-    }
-
-    if (activeDS.currentState != 7)
+    if (activeDS.currentState != 6)
     {
         twRC = TWRC_FAILURE;
         activeDS.twCC = TWCC_SEQERROR;
     }
     else
     {
+        /* Transfer an image from the source to the application */
+        if (SANE_CALL( start_device, NULL ))
+        {
+            activeDS.twCC = TWCC_OPERATIONERROR;
+            return TWRC_FAILURE;
+        }
+
+        if (SANE_CALL( get_params, &activeDS.frame_params ))
+        {
+            WARN("sane_get_parameters failed\n");
+            SANE_CALL( cancel_device, NULL );
+            activeDS.twCC = TWCC_OPERATIONERROR;
+            return TWRC_FAILURE;
+        }
+
         switch (activeDS.frame_params.format)
         {
         case FMT_GRAY:
@@ -496,52 +340,44 @@
             else
             {
                 FIXME("For NATIVE, we support only 1 bit monochrome and 8 bit Grayscale, not %d\n", activeDS.frame_params.depth);
-                SANE_Cancel();
+                SANE_CALL( cancel_device, NULL );
                 activeDS.twCC = TWCC_OPERATIONERROR;
-                activeDS.currentState = 6;
                 return TWRC_FAILURE;
             }
             break;
         case FMT_RGB:
-            if (activeDS.frame_params.depth != 8)
-            {
-                FIXME("For NATIVE, we support only 8 bit per color channel, not %d\n", activeDS.frame_params.depth);
-                SANE_Cancel();
-                activeDS.twCC = TWCC_OPERATIONERROR;
-                activeDS.currentState = 6;
-                return TWRC_FAILURE;
-            }
             break;
         case FMT_OTHER:
             FIXME("For NATIVE, we support only GRAY and RGB\n");
-            SANE_Cancel();
+            SANE_CALL( cancel_device, NULL );
             activeDS.twCC = TWCC_OPERATIONERROR;
-            activeDS.currentState = 6;
-            return TWRC_FAILURE;
-        }
+            return TWRC_FAILURE;
+        }
+
+        TRACE("Acquiring image %dx%dx%d bits (format=%d last=%d bpl=%d) from sane...\n"
+              , activeDS.frame_params.pixels_per_line, activeDS.frame_params.lines,
+              activeDS.frame_params.depth, activeDS.frame_params.format,
+              activeDS.frame_params.last_frame, activeDS.frame_params.bytes_per_line);
 
         dib_bytes_per_line = ((activeDS.frame_params.bytes_per_line + 3) / 4) * 4;
-        y = activeDS.feederEnabled ? activeDS.frame_params.lines * 3/2 : activeDS.frame_params.lines+1;
-        dib_bytes = y * dib_bytes_per_line;
-
-        hDIB = GlobalAlloc(GMEM_MOVEABLE, dib_bytes + sizeof(*header) + color_size);
+        dib_bytes = activeDS.frame_params.lines * dib_bytes_per_line;
+
+        hDIB = GlobalAlloc(GMEM_ZEROINIT, dib_bytes + sizeof(*header) + color_size);
         if (hDIB)
            header = GlobalLock(hDIB);
 
         if (!header)
         {
-            SANE_Cancel();
+            SANE_CALL( cancel_device, NULL );
             activeDS.twCC = TWCC_LOWMEMORY;
-            activeDS.currentState = 6;
             if (hDIB)
                 GlobalFree(hDIB);
             return TWRC_FAILURE;
         }
 
-        memset(header, 0, sizeof(BITMAPINFOHEADER)+color_size);
         header->biSize = sizeof (*header);
         header->biWidth = activeDS.frame_params.pixels_per_line;
-        header->biHeight = -y;
+        header->biHeight = activeDS.frame_params.lines;
         header->biPlanes = 1;
         header->biCompression = BI_RGB;
         switch (activeDS.frame_params.format)
@@ -555,6 +391,7 @@
         case FMT_OTHER:
             break;
         }
+        header->biSizeImage = dib_bytes;
         header->biXPelsPerMeter = 0;
         header->biYPelsPerMeter = 0;
         header->biClrUsed = 0;
@@ -577,120 +414,54 @@
                     colors[i].rgbBlue = colors[i].rgbRed = colors[i].rgbGreen = i;
         }
 
-        y=0;
-        do
+
+        /* Sane returns data in top down order.  Acrobat does best with
+           a bottom up DIB being returned.  */
+        line = p + (activeDS.frame_params.lines - 1) * dib_bytes_per_line;
+        for (i = activeDS.frame_params.lines - 1; i >= 0; i--)
         {
             int retlen;
-            struct read_data_params params = { NULL, activeDS.frame_params.bytes_per_line, &retlen };
-
-            if (y >= -header->biHeight)
-            {
-                TRACE("Data source transfers more lines than expected. Extend DIB to %ld lines\n", (-header->biHeight)+1000);
-                GlobalUnlock(hDIB);
-
-                dib_bytes += 1000 * dib_bytes_per_line;
-
-                if (!GlobalReAlloc(hDIB, dib_bytes + sizeof(*header) + color_size, GMEM_MOVEABLE))
-                {
-                    SANE_Cancel();
-                    activeDS.twCC = TWCC_LOWMEMORY;
-                    activeDS.currentState = 6;
-                    GlobalFree(hDIB);
-                    return TWRC_FAILURE;
-                }
-
-                header = (BITMAPINFOHEADER *) GlobalLock(hDIB);
-                header->biHeight -= 1000;
-
-                p = ((BYTE *) header) + header->biSize + color_size;
-            }
+            struct read_data_params params = { line, activeDS.frame_params.bytes_per_line, &retlen };
 
             activeDS.progressWnd = ScanningDialogBox(activeDS.progressWnd,
-                                                     MulDiv(y, 100, activeDS.frame_params.lines));
-
-            params.buffer =
-              line = p + y * dib_bytes_per_line;
-
-            if (activeDS.frame_params.bytes_per_line & 3)
-            {
-                /* Set padding-bytes at the end of the line buffer to 0 */
-                memset(line+dib_bytes_per_line-4, 0, 4);
-            }
+                    ((activeDS.frame_params.lines - 1 - i) * 100)
+                            /
+                    (activeDS.frame_params.lines - 1));
 
             twRC = SANE_CALL( read_data, &params );
             if (twRC != TWCC_SUCCESS) break;
-            if (retlen < activeDS.frame_params.bytes_per_line)
-            {   /* EOF reached */
-                eof = 1;
-            }
-            else
-            {
-                y++;
-                /* TWAIN: for 24 bit color DIBs, the pixels are stored in BGR order */
-                if (activeDS.frame_params.format == FMT_RGB && activeDS.frame_params.depth == 8)
+            if (retlen < activeDS.frame_params.bytes_per_line) break;
+            /* TWAIN: for 24 bit color DIBs, the pixels are stored in BGR order */
+            if (activeDS.frame_params.format == FMT_RGB && activeDS.frame_params.depth == 8)
+            {
+                pixels = (RGBTRIPLE *) line;
+                for (j = 0; j < activeDS.frame_params.pixels_per_line; ++j)
                 {
-                    pixels = (RGBTRIPLE *) line;
-                    for (j = 0; j < activeDS.frame_params.pixels_per_line; ++j)
-                    {
-                        color_buffer = pixels[j].rgbtRed;
-                        pixels[j].rgbtRed = pixels[j].rgbtBlue;
-                        pixels[j].rgbtBlue = color_buffer;
-                    }
+                    color_buffer = pixels[j].rgbtRed;
+                    pixels[j].rgbtRed = pixels[j].rgbtBlue;
+                    pixels[j].rgbtBlue = color_buffer;
                 }
             }
-        }
-        while (!eof && !activeDS.userCancelled);
-
-        if (twRC != TWCC_SUCCESS || y==0 || activeDS.userCancelled)
-        {
-            WARN("sane_read: %u, reading line %d\n", twRC, y);
-            SANE_Cancel();
+            line -= dib_bytes_per_line;
+        }
+        activeDS.progressWnd = ScanningDialogBox(activeDS.progressWnd, -1);
+
+        GlobalUnlock(hDIB);
+
+        if (twRC != TWCC_SUCCESS)
+        {
+            WARN("sane_read: %u, reading line %d\n", twRC, i);
+            SANE_CALL( cancel_device, NULL );
             activeDS.twCC = TWCC_OPERATIONERROR;
-            GlobalUnlock(hDIB);
             GlobalFree(hDIB);
             return TWRC_FAILURE;
         }
 
-        /* Sane returns data in top down order.  Acrobat does best with
-           a bottom up DIB being returned. Flip image */
-        header->biHeight = y;
-        for (y = 0; y<header->biHeight / 2; y++)
-        {
-            ptop = (DWORD *) (p + dib_bytes_per_line *                   y    );
-            pbot = (DWORD *) (p + dib_bytes_per_line * (header->biHeight-y-1) );
-
-            for (i=0; i<dib_bytes_per_line/4; i++)
-            {
-                tmp = *ptop;
-                *ptop = *pbot;
-                *pbot = tmp;
-                ptop++;
-                pbot++;
-            }
-        }
-
-        /* Shrink the memory block to the size actually transfered by
-         * the sane data source. */
-        header->biSizeImage =
-          dib_bytes = header->biHeight * dib_bytes_per_line;
-
-        GlobalUnlock(hDIB);
-
-        if (!GlobalReAlloc(hDIB, dib_bytes + sizeof(*header) + color_size, GMEM_MOVEABLE))
-        {
-            SANE_Cancel();
-            activeDS.twCC = TWCC_LOWMEMORY;
-            activeDS.currentState = 6;
-            GlobalUnlock(hDIB);
-            GlobalFree(hDIB);
-            return TWRC_FAILURE;
-        }
-
-        activeDS.progressWnd = ScanningDialogBox(activeDS.progressWnd, -1);
-
-        *pHandle = (TW_HANDLE)hDIB;
+        SANE_CALL( cancel_device, NULL );
+        *pHandle = (UINT_PTR)hDIB;
         twRC = TWRC_XFERDONE;
         activeDS.twCC = TWCC_SUCCESS;
+        activeDS.currentState = 7;
     }
     return twRC;
 }
