diff --git a/dlls/wow64win/gdi.c b/dlls/wow64win/gdi.c
--- a/dlls/wow64win/gdi.c
+++ b/dlls/wow64win/gdi.c
@@ -38,6 +38,17 @@
     WORD  bmBitsPixel;
     ULONG bmBits;
 } BITMAP32;
+
+typedef struct
+{
+    DWORD    elpPenStyle;
+    DWORD    elpWidth;
+    UINT     elpBrushStyle;
+    COLORREF elpColor;
+    ULONG    elpHatch;
+    DWORD    elpNumEntries;
+    DWORD    elpStyleEntry[1];
+} EXTLOGPEN32;
 
 typedef struct
 {
@@ -191,13 +202,6 @@
                         x_src, y_src, rop, bk_color, fl );
 }
 
-NTSTATUS WINAPI wow64_NtGdiCancelDC( UINT *args )
-{
-    HDC hdc = get_handle( &args );
-
-    return NtGdiCancelDC( hdc );
-}
-
 NTSTATUS WINAPI wow64_NtGdiCloseFigure( UINT *args )
 {
     HDC hdc = get_handle( &args );
@@ -397,66 +401,6 @@
     HBRUSH brush = get_handle( &args );
 
     return HandleToUlong( NtGdiCreateSolidBrush( color, brush ));
-}
-
-NTSTATUS WINAPI wow64_NtGdiDdDDIAcquireKeyedMutex( UINT *args )
-{
-    struct
-    {
-        D3DKMT_HANDLE hKeyedMutex;
-        UINT64 Key;
-        ULONG pTimeout;
-        UINT64 FenceValue;
-    } *desc32 = get_ptr( &args );
-    D3DKMT_ACQUIREKEYEDMUTEX desc;
-    NTSTATUS status;
-
-    if (!desc32) return STATUS_INVALID_PARAMETER;
-    desc.hKeyedMutex = desc32->hKeyedMutex;
-    desc.Key = desc32->Key;
-    desc.pTimeout = UlongToHandle( desc32->pTimeout );
-    desc.FenceValue = desc32->FenceValue;
-    status = NtGdiDdDDIAcquireKeyedMutex( &desc );
-    desc32->FenceValue = desc.FenceValue;
-
-    return status;
-}
-
-NTSTATUS WINAPI wow64_NtGdiDdDDIAcquireKeyedMutex2( UINT *args )
-{
-    struct
-    {
-        D3DKMT_HANDLE hKeyedMutex;
-        UINT64 Key;
-        ULONG pTimeout;
-        UINT64 FenceValue;
-        ULONG pPrivateRuntimeData;
-        UINT PrivateRuntimeDataSize;
-    } *desc32 = get_ptr( &args );
-    D3DKMT_ACQUIREKEYEDMUTEX2 desc;
-    NTSTATUS status;
-
-    if (!desc32) return STATUS_INVALID_PARAMETER;
-    desc.hKeyedMutex = desc32->hKeyedMutex;
-    desc.Key = desc32->Key;
-    desc.pTimeout = UlongToHandle( desc32->pTimeout );
-    desc.FenceValue = desc32->FenceValue;
-    desc.pPrivateRuntimeData = UlongToHandle( desc32->pPrivateRuntimeData );
-    desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
-    status = NtGdiDdDDIAcquireKeyedMutex2( &desc );
-    desc32->FenceValue = desc.FenceValue;
-
-    return status;
-}
-
-NTSTATUS WINAPI wow64_NtGdiDdDDICheckOcclusion( UINT *args )
-{
-    struct
-    {
-        ULONG hWnd;
-    } *desc32 = get_ptr( &args );
-    D3DKMT_CHECKOCCLUSION desc = {.hWnd = UlongToHandle( desc32->hWnd )};
-    return NtGdiDdDDICheckOcclusion( &desc );
 }
 
 NTSTATUS WINAPI wow64_NtGdiDdDDICheckVidPnExclusiveOwnership( UINT *args )
@@ -531,7 +475,6 @@
 
     desc.hDevice = desc32->hDevice;
     desc.hResource = desc32->hResource;
-    desc.hGlobalShare = desc32->hGlobalShare;
     desc.pPrivateRuntimeData = UlongToPtr( desc32->pPrivateRuntimeData );
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
     if (!desc32->Flags.StandardAllocation)
@@ -547,16 +490,16 @@
         standard.Flags = standard32->Flags;
 
         desc.pStandardAllocation = &standard;
-        desc.PrivateDriverDataSize = desc32->PrivateDriverDataSize;
+        desc.PrivateDriverDataSize = sizeof(standard);
     }
     desc.NumAllocations = desc32->NumAllocations;
-    allocs32 = UlongToPtr( desc32->pAllocationInfo );
     desc.pAllocationInfo = NULL;
     if (desc32->pAllocationInfo && desc32->NumAllocations)
     {
         if (!(desc.pAllocationInfo = Wow64AllocateTemp( desc32->NumAllocations + sizeof(*desc.pAllocationInfo) )))
             return STATUS_NO_MEMORY;
 
+        allocs32 = UlongToPtr( desc32->pAllocationInfo );
         for (i = 0; i < desc32->NumAllocations; i++)
         {
             desc.pAllocationInfo[i].hAllocation = allocs32->hAllocation;
@@ -571,10 +514,8 @@
     desc.hPrivateRuntimeResourceHandle = desc32->hPrivateRuntimeResourceHandle;
 
     status = NtGdiDdDDICreateAllocation( &desc );
-    desc32->hResource = desc.hResource;
-    desc32->hGlobalShare = desc.hGlobalShare;
-    for (i = 0; desc32->pAllocationInfo && i < desc32->NumAllocations; i++)
-        allocs32->hAllocation = desc.pAllocationInfo[i].hAllocation;
+    desc.hResource = desc32->hResource;
+    desc.hGlobalShare = desc32->hGlobalShare;
     return status;
 }
 
@@ -640,7 +581,6 @@
 
     desc.hDevice = desc32->hDevice;
     desc.hResource = desc32->hResource;
-    desc.hGlobalShare = desc32->hGlobalShare;
     desc.pPrivateRuntimeData = UlongToPtr( desc32->pPrivateRuntimeData );
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
     if (!desc32->Flags.StandardAllocation)
@@ -656,16 +596,16 @@
         standard.Flags = standard32->Flags;
 
         desc.pStandardAllocation = &standard;
-        desc.PrivateDriverDataSize = desc32->PrivateDriverDataSize;
+        desc.PrivateDriverDataSize = sizeof(standard);
     }
     desc.NumAllocations = desc32->NumAllocations;
-    allocs32 = UlongToPtr( desc32->pAllocationInfo2 );
     desc.pAllocationInfo2 = NULL;
     if (desc32->pAllocationInfo2 && desc32->NumAllocations)
     {
         if (!(desc.pAllocationInfo2 = Wow64AllocateTemp( desc32->NumAllocations + sizeof(*desc.pAllocationInfo2) )))
             return STATUS_NO_MEMORY;
 
+        allocs32 = UlongToPtr( desc32->pAllocationInfo2 );
         for (i = 0; i < desc32->NumAllocations; i++)
         {
             desc.pAllocationInfo2[i].hAllocation = allocs32->hAllocation;
@@ -681,13 +621,10 @@
     desc.hPrivateRuntimeResourceHandle = desc32->hPrivateRuntimeResourceHandle;
 
     status = NtGdiDdDDICreateAllocation( &desc );
-    desc32->hResource = desc.hResource;
-    desc32->hGlobalShare = desc.hGlobalShare;
+    desc.hResource = desc32->hResource;
+    desc.hGlobalShare = desc32->hGlobalShare;
     for (i = 0; desc32->pAllocationInfo2 && i < desc32->NumAllocations; i++)
-    {
-        allocs32->hAllocation = desc.pAllocationInfo2[i].hAllocation;
         allocs32->GpuVirtualAddress = desc.pAllocationInfo2[i].GpuVirtualAddress;
-    }
     return status;
 }
 
@@ -717,8 +654,6 @@
     desc.Pitch = desc32->Pitch;
     desc.hDeviceDc = UlongToHandle( desc32->hDeviceDc );
     desc.pColorTable = UlongToPtr( desc32->pColorTable );
-    desc.hDc = UlongToHandle( desc32->hDc );
-    desc.hBitmap = UlongToHandle( desc32->hBitmap );
 
     if (!(status = NtGdiDdDDICreateDCFromMemory( &desc )))
     {
@@ -748,7 +683,6 @@
 
     if (!desc32) return STATUS_INVALID_PARAMETER;
     desc.hAdapter = desc32->hAdapter;
-    desc.hDevice = desc32->hDevice;
     desc.Flags = desc32->Flags;
     desc.pCommandBuffer = UlongToPtr( desc32->pCommandBuffer );
     desc.CommandBufferSize = desc32->CommandBufferSize;
@@ -781,17 +715,13 @@
     D3DKMT_CREATEKEYEDMUTEX2 desc;
     NTSTATUS status;
 
-    if (!desc32) return STATUS_INVALID_PARAMETER;
-
     desc.InitialValue = desc32->InitialValue;
     desc.hSharedHandle = desc32->hSharedHandle;
-    desc.hKeyedMutex = desc32->hKeyedMutex;
     desc.pPrivateRuntimeData = ULongToPtr( desc32->pPrivateRuntimeData );
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
     desc.Flags = desc32->Flags;
     status = NtGdiDdDDICreateKeyedMutex2( &desc );
     desc32->hKeyedMutex = desc.hKeyedMutex;
-    desc32->hSharedHandle = desc.hSharedHandle;
     return status;
 }
 
@@ -805,7 +735,6 @@
 {
     D3DKMT_CREATESYNCHRONIZATIONOBJECT2 *desc = get_ptr( &args );
 
-    if (!desc) return STATUS_INVALID_PARAMETER;
     if (desc->Info.Type == D3DDDI_CPU_NOTIFICATION)
     {
         ULONG event = HandleToUlong( desc->Info.CPUNotification.Event );
@@ -956,8 +885,6 @@
 
     if (!desc32) return STATUS_INVALID_PARAMETER;
     desc.pDeviceName = UlongToPtr( desc32->pDeviceName );
-    desc.hAdapter = desc32->hAdapter;
-    desc.AdapterLuid = desc32->AdapterLuid;
 
     if (!(status = NtGdiDdDDIOpenAdapterFromDeviceName( &desc )))
     {
@@ -982,10 +909,6 @@
 
     if (!desc32) return STATUS_INVALID_PARAMETER;
     desc.hDc = UlongToHandle( desc32->hDc );
-    desc.hAdapter = desc32->hAdapter;
-    desc.AdapterLuid = desc32->AdapterLuid;
-    desc.VidPnSourceId = desc32->VidPnSourceId;
-
     if (!(status = NtGdiDdDDIOpenAdapterFromHdc( &desc )))
     {
         desc32->hAdapter = desc.hAdapter;
@@ -1021,7 +944,6 @@
     NTSTATUS status;
 
     desc.hSharedHandle = desc32->hSharedHandle;
-    desc.hKeyedMutex = desc32->hKeyedMutex;
     desc.pPrivateRuntimeData = UlongToPtr( desc32->pPrivateRuntimeData );
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
     status = NtGdiDdDDIOpenKeyedMutex2( &desc );
@@ -1042,31 +964,10 @@
     NTSTATUS status;
 
     desc.hNtHandle = UlongToHandle( desc32->hNtHandle );
-    desc.hKeyedMutex = desc32->hKeyedMutex;
     desc.pPrivateRuntimeData = UlongToPtr( desc32->pPrivateRuntimeData );
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
     status = NtGdiDdDDIOpenKeyedMutexFromNtHandle( &desc );
     desc32->hKeyedMutex = desc.hKeyedMutex;
-    return status;
-}
-
-NTSTATUS WINAPI wow64_NtGdiDdDDIOpenNtHandleFromName( UINT *args )
-{
-    struct
-    {
-        DWORD dwDesiredAccess;
-        ULONG pObjAttrib;
-        ULONG hNtHandle;
-    } *desc32 = get_ptr( &args );
-    D3DKMT_OPENNTHANDLEFROMNAME desc;
-    struct object_attr64 attr;
-    NTSTATUS status;
-
-    desc.dwDesiredAccess = desc32->dwDesiredAccess;
-    desc.pObjAttrib = objattr_32to64( &attr, UlongToPtr( desc32->pObjAttrib ) );
-    desc.hNtHandle = UlongToHandle( desc32->hNtHandle );
-    status = NtGdiDdDDIOpenNtHandleFromName( &desc );
-    desc32->hNtHandle = HandleToUlong( desc.hNtHandle );
     return status;
 }
 
@@ -1099,13 +1000,13 @@
     desc.hDevice = desc32->hDevice;
     desc.hGlobalShare = desc32->hGlobalShare;
     desc.NumAllocations = desc32->NumAllocations;
-    allocs32 = UlongToPtr( desc32->pOpenAllocationInfo );
     desc.pOpenAllocationInfo = NULL;
     if (desc32->pOpenAllocationInfo && desc32->NumAllocations)
     {
         if (!(desc.pOpenAllocationInfo = Wow64AllocateTemp( desc32->NumAllocations + sizeof(*desc.pOpenAllocationInfo) )))
             return STATUS_NO_MEMORY;
 
+        allocs32 = UlongToPtr( desc32->pOpenAllocationInfo );
         for (i = 0; i < desc32->NumAllocations; i++)
         {
             desc.pOpenAllocationInfo[i].hAllocation = allocs32->hAllocation;
@@ -1119,16 +1020,10 @@
     desc.pResourcePrivateDriverData = UlongToPtr( desc32->pResourcePrivateDriverData );
     desc.TotalPrivateDriverDataBufferSize = desc32->TotalPrivateDriverDataBufferSize;
     desc.pTotalPrivateDriverDataBuffer = UlongToPtr( desc32->pTotalPrivateDriverDataBuffer );
-    desc.hResource = desc32->hResource;
 
     status = NtGdiDdDDIOpenResource( &desc );
     desc32->TotalPrivateDriverDataBufferSize = desc.TotalPrivateDriverDataBufferSize;
     desc32->hResource = desc.hResource;
-    for (i = 0; desc32->pOpenAllocationInfo && i < desc32->NumAllocations; i++)
-    {
-        allocs32->hAllocation = desc.pOpenAllocationInfo[i].hAllocation;
-        allocs32->PrivateDriverDataSize = desc.pOpenAllocationInfo[i].PrivateDriverDataSize;
-    }
     return status;
 }
 
@@ -1163,19 +1058,18 @@
     desc.hDevice = desc32->hDevice;
     desc.hGlobalShare = desc32->hGlobalShare;
     desc.NumAllocations = desc32->NumAllocations;
-    allocs32 = UlongToPtr( desc32->pOpenAllocationInfo2 );
     desc.pOpenAllocationInfo2 = NULL;
     if (desc32->pOpenAllocationInfo2 && desc32->NumAllocations)
     {
         if (!(desc.pOpenAllocationInfo2 = Wow64AllocateTemp( desc32->NumAllocations + sizeof(*desc.pOpenAllocationInfo2) )))
             return STATUS_NO_MEMORY;
 
+        allocs32 = UlongToPtr( desc32->pOpenAllocationInfo2 );
         for (i = 0; i < desc32->NumAllocations; i++)
         {
             desc.pOpenAllocationInfo2[i].hAllocation = allocs32->hAllocation;
             desc.pOpenAllocationInfo2[i].pPrivateDriverData = UlongToPtr( allocs32->pPrivateDriverData );
             desc.pOpenAllocationInfo2[i].PrivateDriverDataSize = allocs32->PrivateDriverDataSize;
-            desc.pOpenAllocationInfo2[i].GpuVirtualAddress = allocs32->GpuVirtualAddress;
         }
     }
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
@@ -1184,17 +1078,12 @@
     desc.pResourcePrivateDriverData = UlongToPtr( desc32->pResourcePrivateDriverData );
     desc.TotalPrivateDriverDataBufferSize = desc32->TotalPrivateDriverDataBufferSize;
     desc.pTotalPrivateDriverDataBuffer = UlongToPtr( desc32->pTotalPrivateDriverDataBuffer );
-    desc.hResource = desc32->hResource;
 
     status = NtGdiDdDDIOpenResource2( &desc );
     desc32->TotalPrivateDriverDataBufferSize = desc.TotalPrivateDriverDataBufferSize;
     desc32->hResource = desc.hResource;
     for (i = 0; desc32->pOpenAllocationInfo2 && i < desc32->NumAllocations; i++)
-    {
-        allocs32->hAllocation = desc.pOpenAllocationInfo2[i].hAllocation;
-        allocs32->PrivateDriverDataSize = desc.pOpenAllocationInfo2[i].PrivateDriverDataSize;
         allocs32->GpuVirtualAddress = desc.pOpenAllocationInfo2[i].GpuVirtualAddress;
-    }
     return status;
 }
 
@@ -1233,19 +1122,18 @@
     desc.hDevice = desc32->hDevice;
     desc.hNtHandle = UlongToHandle( desc32->hNtHandle );
     desc.NumAllocations = desc32->NumAllocations;
-    allocs32 = UlongToPtr( desc32->pOpenAllocationInfo2 );
     desc.pOpenAllocationInfo2 = NULL;
     if (desc32->pOpenAllocationInfo2 && desc32->NumAllocations)
     {
         if (!(desc.pOpenAllocationInfo2 = Wow64AllocateTemp( desc32->NumAllocations + sizeof(*desc.pOpenAllocationInfo2) )))
             return STATUS_NO_MEMORY;
 
+        allocs32 = UlongToPtr( desc32->pOpenAllocationInfo2 );
         for (i = 0; i < desc32->NumAllocations; i++)
         {
             desc.pOpenAllocationInfo2[i].hAllocation = allocs32->hAllocation;
             desc.pOpenAllocationInfo2[i].pPrivateDriverData = UlongToPtr( allocs32->pPrivateDriverData );
             desc.pOpenAllocationInfo2[i].PrivateDriverDataSize = allocs32->PrivateDriverDataSize;
-            desc.pOpenAllocationInfo2[i].GpuVirtualAddress = allocs32->GpuVirtualAddress;
         }
     }
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
@@ -1256,23 +1144,14 @@
     desc.pTotalPrivateDriverDataBuffer = UlongToPtr( desc32->pTotalPrivateDriverDataBuffer );
     desc.pKeyedMutexPrivateRuntimeData = UlongToPtr( desc32->pKeyedMutexPrivateRuntimeData );
     desc.KeyedMutexPrivateRuntimeDataSize = desc32->KeyedMutexPrivateRuntimeDataSize;
-    desc.hResource = desc32->hResource;
-    desc.hKeyedMutex = desc32->hKeyedMutex;
-    desc.hSyncObject = desc32->hSyncObject;
 
     status = NtGdiDdDDIOpenResourceFromNtHandle( &desc );
-    desc32->PrivateRuntimeDataSize = desc.PrivateRuntimeDataSize;
-    desc32->ResourcePrivateDriverDataSize = desc.ResourcePrivateDriverDataSize;
     desc32->TotalPrivateDriverDataBufferSize = desc.TotalPrivateDriverDataBufferSize;
     desc32->hResource = desc.hResource;
     desc32->hKeyedMutex = desc.hKeyedMutex;
     desc32->hSyncObject = desc.hSyncObject;
     for (i = 0; desc32->pOpenAllocationInfo2 && i < desc32->NumAllocations; i++)
-    {
-        allocs32->hAllocation = desc.pOpenAllocationInfo2[i].hAllocation;
-        allocs32->PrivateDriverDataSize = desc.pOpenAllocationInfo2[i].PrivateDriverDataSize;
         allocs32->GpuVirtualAddress = desc.pOpenAllocationInfo2[i].GpuVirtualAddress;
-    }
     return status;
 }
 
@@ -1287,7 +1166,6 @@
     NTSTATUS status;
 
     desc.hNtHandle = UlongToHandle( desc32->hNtHandle );
-    desc.hSyncObject = desc32->hSyncObject;
     status = NtGdiDdDDIOpenSyncObjectFromNtHandle( &desc );
     desc32->hSyncObject = desc.hSyncObject;
     return status;
@@ -1318,10 +1196,7 @@
     desc.hNtHandle = ULongToHandle( desc32->hNtHandle );
     desc.hDevice = desc32->hDevice;
     desc.Flags = desc32->Flags;
-    desc.hSyncObject = desc32->hSyncObject;
     desc.MonitoredFence.EngineAffinity = desc32->MonitoredFence.EngineAffinity;
-    desc.MonitoredFence.FenceValueCPUVirtualAddress = UlongToPtr( desc32->MonitoredFence.FenceValueCPUVirtualAddress );
-    desc.MonitoredFence.FenceValueGPUVirtualAddress = desc32->MonitoredFence.FenceValueGPUVirtualAddress;
 
     status = NtGdiDdDDIOpenSyncObjectFromNtHandle2( &desc );
     desc32->MonitoredFence.FenceValueCPUVirtualAddress = PtrToUlong( desc.MonitoredFence.FenceValueCPUVirtualAddress );
@@ -1344,7 +1219,6 @@
 
     desc.dwDesiredAccess = desc32->dwDesiredAccess;
     desc.pObjAttrib = objattr_32to64( &attr, UlongToPtr( desc32->pObjAttrib ) );
-    desc.hNtHandle = UlongToHandle( desc32->hNtHandle );
     status = NtGdiDdDDIOpenSyncObjectNtHandleFromName( &desc );
     desc32->hNtHandle = HandleToUlong( desc.hNtHandle );
     return status;
@@ -1396,9 +1270,6 @@
     desc.hGlobalShare = desc32->hGlobalShare;
     desc.pPrivateRuntimeData = UlongToPtr( desc32->pPrivateRuntimeData );
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
-    desc.TotalPrivateDriverDataSize = desc32->TotalPrivateDriverDataSize;
-    desc.ResourcePrivateDriverDataSize = desc32->ResourcePrivateDriverDataSize;
-    desc.NumAllocations = desc32->NumAllocations;
     status = NtGdiDdDDIQueryResourceInfo( &desc );
     desc32->PrivateRuntimeDataSize = desc.PrivateRuntimeDataSize;
     desc32->TotalPrivateDriverDataSize = desc.TotalPrivateDriverDataSize;
@@ -1426,9 +1297,6 @@
     desc.hNtHandle = UlongToHandle( desc32->hNtHandle );
     desc.pPrivateRuntimeData = UlongToPtr( desc32->pPrivateRuntimeData );
     desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
-    desc.TotalPrivateDriverDataSize = desc32->TotalPrivateDriverDataSize;
-    desc.ResourcePrivateDriverDataSize = desc32->ResourcePrivateDriverDataSize;
-    desc.NumAllocations = desc32->NumAllocations;
     status = NtGdiDdDDIQueryResourceInfoFromNtHandle( &desc );
     desc32->PrivateRuntimeDataSize = desc.PrivateRuntimeDataSize;
     desc32->TotalPrivateDriverDataSize = desc.TotalPrivateDriverDataSize;
@@ -1465,10 +1333,6 @@
     desc.hAdapter = desc32->hAdapter;
     desc.MemorySegmentGroup = desc32->MemorySegmentGroup;
     desc.PhysicalAdapterIndex = desc32->PhysicalAdapterIndex;
-    desc.Budget = desc32->Budget;
-    desc.CurrentUsage = desc32->CurrentUsage;
-    desc.CurrentReservation = desc32->CurrentReservation;
-    desc.AvailableForReservation = desc32->AvailableForReservation;
 
     if (!(status = NtGdiDdDDIQueryVideoMemoryInfo( &desc )))
     {
@@ -1480,36 +1344,6 @@
     return status;
 }
 
-NTSTATUS WINAPI wow64_NtGdiDdDDIReleaseKeyedMutex( UINT *args )
-{
-    D3DKMT_RELEASEKEYEDMUTEX *desc = get_ptr( &args );
-    return NtGdiDdDDIReleaseKeyedMutex( desc );
-}
-
-NTSTATUS WINAPI wow64_NtGdiDdDDIReleaseKeyedMutex2( UINT *args )
-{
-    struct
-    {
-        D3DKMT_HANDLE hKeyedMutex;
-        UINT64 Key;
-        UINT64 FenceValue;
-        ULONG pPrivateRuntimeData;
-        UINT PrivateRuntimeDataSize;
-    } *desc32 = get_ptr( &args );
-    D3DKMT_RELEASEKEYEDMUTEX2 desc;
-    NTSTATUS status;
-
-    if (!desc32) return STATUS_INVALID_PARAMETER;
-    desc.hKeyedMutex = desc32->hKeyedMutex;
-    desc.Key = desc32->Key;
-    desc.FenceValue = desc32->FenceValue;
-    desc.pPrivateRuntimeData = UlongToHandle( desc32->pPrivateRuntimeData );
-    desc.PrivateRuntimeDataSize = desc32->PrivateRuntimeDataSize;
-    status = NtGdiDdDDIReleaseKeyedMutex2( &desc );
-
-    return status;
-}
-
 NTSTATUS WINAPI wow64_NtGdiDdDDISetQueuedLimit( UINT *args )
 {
     D3DKMT_SETQUEUEDLIMIT *desc = get_ptr( &args );
@@ -1546,58 +1380,13 @@
     ULONG *handle_ptr = get_ptr( &args );
 
     struct object_attr64 attr;
-    HANDLE handle = UlongToHandle( *handle_ptr );
+    HANDLE handle = 0;
     NTSTATUS status;
 
+    *handle_ptr = 0;
     status = NtGdiDdDDIShareObjects( count, handles, objattr_32to64( &attr, attr32 ), access, &handle );
     *handle_ptr = HandleToULong( handle );
     return status;
-}
-
-NTSTATUS WINAPI wow64_NtGdiDdDDISignalSynchronizationObjectFromCpu( UINT *args )
-{
-    const struct
-    {
-        D3DKMT_HANDLE hDevice;
-        UINT ObjectCount;
-        ULONG ObjectHandleArray;
-        ULONG FenceValueArray;
-        D3DDDICB_SIGNALFLAGS Flags;
-    } *desc32 = get_ptr( &args );
-    D3DKMT_SIGNALSYNCHRONIZATIONOBJECTFROMCPU desc;
-
-    if (!desc32) return STATUS_INVALID_PARAMETER;
-    desc.hDevice = desc32->hDevice;
-    desc.ObjectCount = desc32->ObjectCount;
-    desc.ObjectHandleArray = UlongToPtr( desc32->ObjectHandleArray );
-    desc.FenceValueArray = UlongToPtr( desc32->FenceValueArray );
-    desc.Flags = desc32->Flags;
-
-    return NtGdiDdDDISignalSynchronizationObjectFromCpu( &desc );
-}
-
-NTSTATUS WINAPI wow64_NtGdiDdDDIWaitForSynchronizationObjectFromCpu( UINT *args )
-{
-    const struct
-    {
-        D3DKMT_HANDLE hDevice;
-        UINT ObjectCount;
-        ULONG ObjectHandleArray;
-        ULONG FenceValueArray;
-        ULONG hAsyncEvent;
-        D3DDDI_WAITFORSYNCHRONIZATIONOBJECTFROMCPU_FLAGS Flags;
-    } *desc32 = get_ptr( &args );
-    D3DKMT_WAITFORSYNCHRONIZATIONOBJECTFROMCPU desc;
-
-    if (!desc32) return STATUS_INVALID_PARAMETER;
-    desc.hDevice = desc32->hDevice;
-    desc.ObjectCount = desc32->ObjectCount;
-    desc.ObjectHandleArray = UlongToPtr( desc32->ObjectHandleArray );
-    desc.FenceValueArray = UlongToPtr( desc32->FenceValueArray );
-    desc.hAsyncEvent = UlongToHandle( desc32->hAsyncEvent );
-    desc.Flags = desc32->Flags;
-
-    return NtGdiDdDDIWaitForSynchronizationObjectFromCpu( &desc );
 }
 
 NTSTATUS WINAPI wow64_NtGdiDeleteClientObj( UINT *args )
@@ -2148,14 +1937,6 @@
     return NtGdiGetKerningPairs( hdc, count, kern_pair );
 }
 
-NTSTATUS WINAPI wow64_NtGdiGetMiterLimit( UINT *args )
-{
-    HDC hdc = get_handle( &args );
-    FLOAT *limit = get_ptr( &args );
-
-    return NtGdiGetMiterLimit( hdc, limit );
-}
-
 NTSTATUS WINAPI wow64_NtGdiGetNearestColor( UINT *args )
 {
     HDC hdc = get_handle( &args );
@@ -2363,17 +2144,6 @@
     INT y = get_ulong( &args );
 
     return NtGdiLineTo( hdc, x, y );
-}
-
-NTSTATUS WINAPI wow64_NtGdiMakeFontDir( UINT *args )
-{
-    DWORD embed = get_ulong( &args );
-    BYTE *buffer = get_ptr( &args );
-    UINT size = get_ulong( &args );
-    WCHAR *path = get_ptr( &args );
-    UINT len = get_ulong( &args );
-
-    return NtGdiMakeFontDir( embed, buffer, size, path, len );
 }
 
 NTSTATUS WINAPI wow64_NtGdiMaskBlt( UINT *args )
@@ -2848,15 +2618,6 @@
     return NtGdiSetMetaRgn( hdc );
 }
 
-NTSTATUS WINAPI wow64_NtGdiSetMiterLimit( UINT *args )
-{
-    HDC hdc = get_handle( &args );
-    DWORD limit = get_ulong( &args );
-    FLOAT *old_limit = get_ptr( &args );
-
-    return NtGdiSetMiterLimit( hdc, limit, old_limit );
-}
-
 NTSTATUS WINAPI wow64_NtGdiSetPixel( UINT *args )
 {
     HDC hdc = get_handle( &args );
@@ -3058,3 +2819,23 @@
 
     return NtGdiWidenPath( hdc );
 }
+
+NTSTATUS WINAPI wow64___wine_get_icm_profile( UINT *args )
+{
+    HDC hdc = get_handle( &args );
+    BOOL allow_default = get_ulong( &args );
+    DWORD *size = get_ptr( &args );
+    WCHAR *filename = get_ptr( &args );
+
+    return __wine_get_icm_profile( hdc, allow_default, size, filename );
+}
+
+NTSTATUS WINAPI wow64___wine_get_file_outline_text_metric( UINT *args )
+{
+    const WCHAR *path = get_ptr( &args );
+    TEXTMETRICW *otm = get_ptr( &args );
+    UINT *em_square = get_ptr( &args );
+    WCHAR *face_name = get_ptr( &args );
+
+    return __wine_get_file_outline_text_metric( path, otm, em_square, face_name );
+}
