diff --git a/dlls/ntdll/unix/env.c b/dlls/ntdll/unix/env.c
--- a/dlls/ntdll/unix/env.c
+++ b/dlls/ntdll/unix/env.c
@@ -63,6 +63,7 @@
 #include "error.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(environ);
+WINE_DECLARE_DEBUG_CHANNEL(nls);
 
 PEB *peb = NULL;
 WOW_PEB *wow_peb = NULL;
@@ -78,8 +79,6 @@
 
 static LCID user_lcid, system_lcid;
 static LANGID user_ui_language;
-static WCHAR *nt_build_dir;
-static WCHAR *nt_data_dir;
 
 static char system_locale[LOCALE_NAME_MAX_LENGTH];
 static char user_locale[LOCALE_NAME_MAX_LENGTH];
@@ -270,7 +269,6 @@
     { "ISO88599", 28599 },
     { "KOI8R", 20866 },
     { "KOI8U", 21866 },
-    { "SHIFTJIS", 932 },
     { "SJIS", 932 },
     { "TIS620", 28601 },
     { "UTF8", CP_UTF8 }
@@ -278,13 +276,14 @@
 
 static void init_unix_codepage(void)
 {
+    const char *name, *ctype;
     char charset_name[16];
-    const char *name;
     size_t i, j;
     int min = 0, max = ARRAY_SIZE(charset_names) - 1;
 
-    setlocale( LC_CTYPE, "" );
-    if (!(name = nl_langinfo( CODESET ))) return;
+    if (!(ctype = setlocale( LC_CTYPE, "" ))) name = "UTF-8";
+    else if (!(name = nl_langinfo( CODESET ))) return;
+    TRACE_(nls)( "Unix LC_CTYPE %s, using %s codeset\n", debugstr_a(ctype), debugstr_a(name) );
 
     /* remove punctuation characters from charset name */
     for (i = j = 0; name[i] && j < sizeof(charset_name)-1; i++)
@@ -320,13 +319,21 @@
 #endif  /* __APPLE__ || __ANDROID__ */
 
 
+static inline SIZE_T get_env_length( const WCHAR *env )
+{
+    const WCHAR *end = env;
+    while (*end) end += wcslen(end) + 1;
+    return end + 1 - env;
+}
+
+
 #define STARTS_WITH(var,str) (!strncmp( var, str, sizeof(str) - 1 ))
 
 /***********************************************************************
  *           is_special_env_var
  *
  * Check if an environment variable needs to be handled specially when
- * imported from the Unix environment (i.e. prefixed with "WINE_HOST_").
+ * passed through the Unix environment (i.e. prefixed with "WINE").
  */
 static BOOL is_special_env_var( const char *var )
 {
@@ -335,29 +342,30 @@
             STARTS_WITH( var, "HOME=" ) ||
             STARTS_WITH( var, "TEMP=" ) ||
             STARTS_WITH( var, "TMP=" ) ||
-            STARTS_WITH( var, "TMPDIR=" ) ||
-            STARTS_WITH( var, "XDG_" ));
-}
-
-
-/***********************************************************************
- *           is_ignored_env_var
- *
- * Check if an environment variable needs to be skipped when importing
- * from the Unix environment, either because it would confuse Windows
- * apps, or make the Windows environment grow too large.
- */
-static BOOL is_ignored_env_var( const char *var )
-{
-    return (STARTS_WITH( var, "NIXPKGS_" ) ||
             STARTS_WITH( var, "QT_" ) ||
             STARTS_WITH( var, "SDL_AUDIODRIVER=" ) ||
-            STARTS_WITH( var, "SDL_AUDIO_DRIVER=" ) ||
-            STARTS_WITH( var, "SDL_VIDEODRIVER=" ) ||
-            STARTS_WITH( var, "SDL_VIDEO_DRIVER=" ) ||
-            STARTS_WITH( var, "VK_" ));
-}
-
+            STARTS_WITH( var, "VK_" ) ||
+            STARTS_WITH( var, "XR_" ) ||
+            STARTS_WITH( var, "XDG_SESSION_TYPE=" ));
+}
+
+/* check if an environment variable changes dynamically in every new process */
+static BOOL is_dynamic_env_var( const char *var )
+{
+    return (STARTS_WITH( var, "WINEDLLOVERRIDES=" ) ||
+            STARTS_WITH( var, "WINEDATADIR=" ) ||
+            STARTS_WITH( var, "WINEHOMEDIR=" ) ||
+            STARTS_WITH( var, "WINEBUILDDIR=" ) ||
+            STARTS_WITH( var, "WINECONFIGDIR=" ) ||
+            STARTS_WITH( var, "WINELOADER=" ) ||
+            STARTS_WITH( var, "WINEDLLDIR" ) ||
+            STARTS_WITH( var, "WINEUNIXCP=" ) ||
+            STARTS_WITH( var, "WINEUSERLOCALE=" ) ||
+            STARTS_WITH( var, "WINEUSERNAME=" ) ||
+            STARTS_WITH( var, "WINEPRELOADRESERVE=" ) ||
+            STARTS_WITH( var, "WINELOADERNOEXEC=" ) ||
+            STARTS_WITH( var, "WINESERVERSOCKET=" ));
+}
 
 /******************************************************************
  *      ntdll_umbstowcs  (ntdll.so)
@@ -462,18 +470,86 @@
 /***********************************************************************
  *           ntdll_get_build_dir  (ntdll.so)
  */
-const WCHAR *ntdll_get_build_dir(void)
-{
-    return nt_build_dir;
+const char *ntdll_get_build_dir(void)
+{
+    return build_dir;
 }
 
 
 /***********************************************************************
  *           ntdll_get_data_dir  (ntdll.so)
  */
-const WCHAR *ntdll_get_data_dir(void)
-{
-    return nt_data_dir;
+const char *ntdll_get_data_dir(void)
+{
+    return data_dir;
+}
+
+
+/***********************************************************************
+ *           build_envp
+ *
+ * Build the environment of a new child process.
+ */
+char **build_envp( const WCHAR *envW )
+{
+    static const char * const unix_vars[] = { "PATH", "TEMP", "TMP", "HOME" };
+    char **envp;
+    char *env, *p;
+    int count = 1, length, lenW;
+    unsigned int i;
+
+    lenW = get_env_length( envW );
+    if (!(env = malloc( lenW * 3 ))) return NULL;
+    length = ntdll_wcstoumbs( envW, lenW, env, lenW * 3, FALSE );
+
+    for (p = env; *p; p += strlen(p) + 1, count++)
+    {
+        if (is_dynamic_env_var( p )) continue;
+        if (is_special_env_var( p )) length += 4; /* prefix it with "WINE" */
+    }
+
+    for (i = 0; i < ARRAY_SIZE( unix_vars ); i++)
+    {
+        if (!(p = getenv(unix_vars[i]))) continue;
+        length += strlen(unix_vars[i]) + strlen(p) + 2;
+        count++;
+    }
+
+    if ((envp = malloc( count * sizeof(*envp) + length )))
+    {
+        char **envptr = envp;
+        char *dst = (char *)(envp + count);
+
+        /* some variables must not be modified, so we get them directly from the unix env */
+        for (i = 0; i < ARRAY_SIZE( unix_vars ); i++)
+        {
+            if (!(p = getenv( unix_vars[i] ))) continue;
+            *envptr++ = strcpy( dst, unix_vars[i] );
+            strcat( dst, "=" );
+            strcat( dst, p );
+            dst += strlen(dst) + 1;
+        }
+
+        /* now put the Windows environment strings */
+        for (p = env; *p; p += strlen(p) + 1)
+        {
+            if (*p == '=') continue;  /* skip drive curdirs, this crashes some unix apps */
+            if (is_dynamic_env_var( p )) continue;
+            if (is_special_env_var( p ))  /* prefix it with "WINE" */
+            {
+                *envptr++ = strcpy( dst, "WINE" );
+                strcat( dst, p );
+            }
+            else
+            {
+                *envptr++ = strcpy( dst, p );
+            }
+            dst += strlen(dst) + 1;
+        }
+        *envptr = 0;
+    }
+    free( env );
+    return envp;
 }
 
 
@@ -726,13 +802,29 @@
 static void init_locale(void)
 {
     struct locale_nls_header *header;
+    const char *all, *ctype, *messages;
     const NLS_LOCALE_HEADER *locale_table;
     const NLS_LOCALE_DATA *locale;
     char *p;
 
-    setlocale( LC_ALL, "" );
-    if (!unix_to_win_locale( setlocale( LC_CTYPE, NULL ), system_locale )) system_locale[0] = 0;
-    if (!unix_to_win_locale( setlocale( LC_MESSAGES, NULL ), user_locale )) user_locale[0] = 0;
+    if (!(all = setlocale( LC_ALL, "" )) && (all = getenv( "LC_ALL" )))
+        FIXME_(nls)( "Failed to set LC_ALL to %s, is the locale supported?\n", debugstr_a(all) );
+    if (!(ctype = setlocale( LC_CTYPE, "" )) && (ctype = getenv( "LC_CTYPE" )))
+        FIXME_(nls)( "Failed to set LC_CTYPE to %s, is the locale supported?\n", debugstr_a(ctype) );
+    if (!(messages = setlocale( LC_MESSAGES, "" )) && (messages = getenv( "LC_MESSAGES" )))
+        FIXME_(nls)( "Failed to set LC_MESSAGES to %s, is the locale supported?\n", debugstr_a(messages) );
+
+    if (!unix_to_win_locale( ctype, system_locale )) system_locale[0] = 0;
+    TRACE_(nls)( "Unix LC_CTYPE is %s, setting system locale to %s\n", debugstr_a(ctype), debugstr_a(user_locale) );
+
+    if (main_argc > 1 && strstr(main_argv[1], "start_protected_game.exe"))
+    {
+        FIXME( "HACK setting EN locale.\n" );
+        messages = "en-US";
+    }
+
+    if (!unix_to_win_locale( messages, user_locale )) user_locale[0] = 0;
+    TRACE_(nls)( "Unix LC_MESSAGES is %s, user system locale to %s\n", debugstr_a(messages), debugstr_a(user_locale) );
 
 #ifdef __APPLE__
     if (!system_locale[0])
@@ -836,20 +928,6 @@
 }
 
 
-/* check if a WINE_HOST_ prefixed variable already exists in the environment */
-static BOOL host_var_exists( const char *name )
-{
-    char *end = strchr( name, '=' );
-
-    if (!end) return FALSE;
-    for (char **e = environ; *e; e++)
-    {
-        if (!STARTS_WITH( *e, "WINE_HOST_" )) continue;
-        if (!strncmp( *e + 10, name, end + 1 - name )) return TRUE;
-    }
-    return FALSE;
-}
-
 static const char overrides_help_message[] =
     "Syntax:\n"
     "  WINEDLLOVERRIDES=\"entry;entry;entry...\"\n"
@@ -873,7 +951,7 @@
 
     /* estimate needed size */
     *size = 1;
-    for (e = environ; *e; e++) *size += strlen(*e) + 6;
+    for (e = environ; *e; e++) *size += strlen(*e) + 1;
 
     env = malloc( *size * sizeof(WCHAR) );
     ptr = env;
@@ -883,24 +961,18 @@
         char *str = *e;
 
         /* skip Unix special variables and use the Wine variants instead */
-        if (STARTS_WITH( str, "WINE" ))
-        {
-            if (is_special_env_var( str + 4 ) || is_ignored_env_var( str + 4 )) str += 4;
+        if (!strncmp( str, "WINE", 4 ))
+        {
+            if (is_special_env_var( str + 4 )) str += 4;
             else if (!strcmp( str, "WINEDLLOVERRIDES=help" ))
             {
                 MESSAGE( overrides_help_message );
                 exit(0);
             }
         }
-        else if (is_ignored_env_var( str )) continue;
-        else if (host_var_exists( str )) continue;
-        else if (is_special_env_var( str )) /* prefix it with WINE_HOST_ */
-        {
-            static const WCHAR hostW[] = {'W','I','N','E','_','H','O','S','T','_'};
-            memcpy( ptr, hostW, sizeof(hostW) );
-            ptr += ARRAY_SIZE(hostW);
-        }
-
+        else if (is_special_env_var( str )) continue;  /* skip it */
+
+        if (is_dynamic_env_var( str )) continue;
         ptr += ntdll_umbstowcs( str, strlen(str) + 1, ptr, end - ptr );
     }
     *pos = ptr - env;
@@ -986,7 +1058,7 @@
 {
     WCHAR *nt_name = NULL;
 
-    if (path && unix_to_nt_file_name( path, &nt_name, FILE_OPEN )) return;
+    if (path && unix_to_nt_file_name( path, &nt_name )) return;
     append_envW( env, pos, size, name, nt_name );
     free( nt_name );
 }
@@ -1002,7 +1074,7 @@
     {
         WCHAR *nt_name = NULL;
 
-        if (!unix_to_nt_file_name( system_dll_paths[i], &nt_name, FILE_OPEN ))
+        if (!unix_to_nt_file_name( system_dll_paths[i], &nt_name ))
         {
             size_t len = wcslen( nt_name );
             path = realloc( path, (path_len + len + 1) * sizeof(WCHAR) );
@@ -1028,17 +1100,16 @@
  */
 static void add_dynamic_environment( WCHAR **env, SIZE_T *pos, SIZE_T *size )
 {
+    const char *overrides = getenv( "WINEDLLOVERRIDES" );
+    const char *wineloader = getenv( "WINELOADER" );
+    const char *var;
     unsigned int i;
     char str[22];
 
-    if (build_dir) unix_to_nt_file_name( build_dir, &nt_build_dir, FILE_OPEN );
-    if (data_dir) unix_to_nt_file_name( data_dir, &nt_data_dir, FILE_OPEN );
-
-    append_envW( env, pos, size, "WINEBUILDDIR", nt_build_dir );
-    append_envW( env, pos, size, "WINEDATADIR", nt_data_dir );
+    add_path_var( env, pos, size, "WINEDATADIR", data_dir );
     add_path_var( env, pos, size, "WINEHOMEDIR", home_dir );
+    add_path_var( env, pos, size, "WINEBUILDDIR", build_dir );
     add_path_var( env, pos, size, "WINECONFIGDIR", config_dir );
-    add_path_var( env, pos, size, "WINELOADER", wineloader );
     for (i = 0; dll_paths[i]; i++)
     {
         snprintf( str, sizeof(str), "WINEDLLDIR%u", i );
@@ -1047,7 +1118,9 @@
     snprintf( str, sizeof(str), "WINEDLLDIR%u", i );
     append_envW( env, pos, size, str, NULL );
     add_system_dll_path_var( env, pos, size );
+    append_envA( env, pos, size, "WINELOADER", wineloader );
     append_envA( env, pos, size, "WINEUSERNAME", user_name );
+    append_envA( env, pos, size, "WINEDLLOVERRIDES", overrides );
     if (unix_cp.CodePage != CP_UTF8)
     {
         snprintf( str, sizeof(str), "%u", unix_cp.CodePage );
@@ -1057,6 +1130,9 @@
     append_envA( env, pos, size, "WINEUSERLOCALE", user_locale );
     append_envA( env, pos, size, "SystemDrive", "C:" );
     append_envA( env, pos, size, "SystemRoot", "C:\\windows" );
+
+    if ((var = getenv( "VR_CONFIG_PATH" ))) add_path_var( env, pos, size, "VR_CONFIG_PATH", var );
+    if ((var = getenv( "VR_LOG_PATH" ))) add_path_var( env, pos, size, "VR_LOG_PATH", var );
 }
 
 
@@ -1331,6 +1407,27 @@
     }
 }
 
+static void get_std_handle( int fd, unsigned int access, unsigned int attributes, HANDLE *handle )
+{
+    IO_STATUS_BLOCK io;
+    FILE_POSITION_INFORMATION pos_info;
+    FILE_NAME_INFORMATION name_info;
+    NTSTATUS status;
+
+    wine_server_fd_to_handle( fd, access, attributes, handle );
+    if (!*handle) return;
+
+    /* Python checks if a file is seekable and if so expects the file name to be gettable from handle. */
+    if (NtQueryInformationFile( *handle, &io, &pos_info, sizeof(pos_info), FilePositionInformation ))
+        return;
+
+    TRACE("handle for fd %d is seekable.\n", fd);
+    if (!(status = NtQueryInformationFile( *handle, &io, &name_info, sizeof(name_info), FileNameInformation ))
+          || status == STATUS_BUFFER_OVERFLOW) return;
+    TRACE("closing handle for fd %d.\n", fd);
+    NtClose( *handle );
+    *handle = NULL;
+}
 
 /*************************************************************************
  *		get_initial_console
@@ -1341,9 +1438,9 @@
 {
     int output_fd = -1;
 
-    wine_server_fd_to_handle( 0, GENERIC_READ|SYNCHRONIZE,  OBJ_INHERIT, &params->hStdInput );
-    wine_server_fd_to_handle( 1, GENERIC_WRITE|SYNCHRONIZE, OBJ_INHERIT, &params->hStdOutput );
-    wine_server_fd_to_handle( 2, GENERIC_WRITE|SYNCHRONIZE, OBJ_INHERIT, &params->hStdError );
+    get_std_handle( 0, GENERIC_READ|SYNCHRONIZE,  OBJ_INHERIT, &params->hStdInput );
+    get_std_handle( 1, GENERIC_WRITE|SYNCHRONIZE, OBJ_INHERIT, &params->hStdOutput );
+    get_std_handle( 2, GENERIC_WRITE|SYNCHRONIZE, OBJ_INHERIT, &params->hStdError );
 
     if (main_image_info.SubSystemType != IMAGE_SUBSYSTEM_WINDOWS_CUI)
         return;
@@ -1421,7 +1518,7 @@
 
     if (pwd)
     {
-        if (!unix_to_nt_file_name( pwd, &ret, FILE_OPEN ))
+        if (!unix_to_nt_file_name( pwd, &ret ))
         {
             ULONG len = wcslen( ret );
             if (len && ret[len - 1] != '\\')
@@ -1565,7 +1662,6 @@
         return;
     }
 
-    virtual_init_user_shared_data();
     env[size] = 0;
     params.Flags           = PROCESS_PARAMS_FLAG_NORMALIZED;
     params.Environment     = env;
@@ -1584,9 +1680,18 @@
 
     wine_server_fd_to_handle( 2, GENERIC_WRITE | SYNCHRONIZE, OBJ_INHERIT, &params.hStdError );
 
-    status = NtCreateUserProcess( &process, &thread, PROCESS_ALL_ACCESS, THREAD_ALL_ACCESS,
-                                  NULL, NULL, 0, THREAD_CREATE_FLAGS_CREATE_SUSPENDED, &params,
-                                  &create_info, &ps_attr );
+    if (NtCurrentTeb64() && !NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR])
+    {
+        NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR] = TRUE;
+        status = NtCreateUserProcess( &process, &thread, PROCESS_ALL_ACCESS, THREAD_ALL_ACCESS,
+                                      NULL, NULL, 0, THREAD_CREATE_FLAGS_CREATE_SUSPENDED, &params,
+                                      &create_info, &ps_attr );
+        NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR] = FALSE;
+    }
+    else
+        status = NtCreateUserProcess( &process, &thread, PROCESS_ALL_ACCESS, THREAD_ALL_ACCESS,
+                                      NULL, NULL, 0, THREAD_CREATE_FLAGS_CREATE_SUSPENDED, &params,
+                                      &create_info, &ps_attr );
     NtClose( params.hStdError );
 
     if (status)
@@ -1601,7 +1706,7 @@
 
 wait:
     timeout.QuadPart = (ULONGLONG)5 * 60 * 1000 * -10000;
-    if (NtWaitForMultipleObjects( count, handles, WaitAny, FALSE, &timeout ) == WAIT_TIMEOUT)
+    if (NtWaitForMultipleObjects( count, handles, TRUE, FALSE, &timeout ) == WAIT_TIMEOUT)
         ERR( "boot event wait timed out\n" );
     while (count) NtClose( handles[--count] );
 }
@@ -1621,30 +1726,6 @@
 static inline void put_unicode_string( WCHAR *src, WCHAR **dst, UNICODE_STRING *str )
 {
     copy_unicode_string( &src, dst, str, wcslen(src) * sizeof(WCHAR) );
-}
-
-static void copy_dos_path_string( WCHAR **src, WCHAR **dst, UNICODE_STRING *str,
-                                  UNICODE_STRING *nt_str, UINT len )
-{
-    /* copy the original string into nt_str */
-    nt_str->Buffer = malloc( len + sizeof(WCHAR) );
-    memcpy( nt_str->Buffer, *src, len );
-    nt_str->Buffer[len / sizeof(WCHAR)] = 0;
-    nt_str->Length = len;
-    nt_str->MaximumLength = len + sizeof(WCHAR);
-
-    if (len > 5 * sizeof(WCHAR) && (*src)[5] == ':') /* skip the \??\ prefix */
-    {
-        *src += 4;
-        len -= 4 * sizeof(WCHAR);
-        copy_unicode_string( src, dst, str, len );
-    }
-    else
-    {
-        WCHAR *ptr = *dst;
-        copy_unicode_string( src, dst, str, len );
-        ptr[1] = '\\'; /* change \??\ to \\?\ */
-    }
 }
 
 static inline WCHAR *get_dos_path( WCHAR *nt_path )
@@ -1837,12 +1918,20 @@
     peb->ImageSubSystemMinorVersion = main_image_info.MinorSubsystemVersion;
 
 #ifdef _WIN64
-    if (!is_machine_64bit( main_image_info.Machine ))
-    {
+    switch (main_image_info.Machine)
+    {
+    case IMAGE_FILE_MACHINE_I386:
+    case IMAGE_FILE_MACHINE_ARMNT:
         NtCurrentTeb()->WowTebOffset = teb_offset;
         NtCurrentTeb()->Tib.ExceptionList = (void *)((char *)NtCurrentTeb() + teb_offset);
         wow_peb = (PEB32 *)((char *)peb + page_size);
         set_thread_id( NtCurrentTeb(), GetCurrentProcessId(), GetCurrentThreadId() );
+        ERR( "starting %s in experimental wow64 mode\n", debugstr_us(&params->ImagePathName) );
+        break;
+    case IMAGE_FILE_MACHINE_AMD64:
+        if (main_image_info.Machine == current_machine) break;
+        ERR( "starting %s in experimental ARM64EC mode\n", debugstr_us(&params->ImagePathName) );
+        break;
     }
 #endif
 
@@ -1885,13 +1974,10 @@
     static const WCHAR pathW[] = {'P','A','T','H'};
     RTL_USER_PROCESS_PARAMETERS *params = NULL;
     SIZE_T size, env_pos, env_size;
-    WCHAR *dst, *cmdline, *path, *bootstrap;
+    WCHAR *dst, *image, *cmdline, *path, *bootstrap;
     WCHAR *env = get_initial_environment( &env_pos, &env_size );
     WCHAR *curdir = get_initial_directory();
-    UNICODE_STRING nt_name;
     NTSTATUS status;
-
-    if (NtCurrentTeb64()) NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR] = TRUE;
 
     /* store the initial PATH value */
     path = get_env_var( env, env_pos, pathW, 4 );
@@ -1912,17 +1998,8 @@
     add_registry_environment( &env, &env_pos, &env_size );
     env[env_pos++] = 0;
 
-    get_full_path( main_argv[1], curdir, &nt_name );
-    status = load_main_exe( &nt_name, 0, module );
-    /* fail only if the file contained an explicit path */
-    if (status == STATUS_DLL_NOT_FOUND &&
-        (strpbrk( main_argv[1], "/\\" ) || (main_argv[1][0] && main_argv[1][1] == ':')))
-    {
-        MESSAGE( "wine: failed to open %s\n", debugstr_a(main_argv[1]) );
-        NtTerminateProcess( GetCurrentProcess(), status );
-    }
-
-    if (NT_SUCCESS(status))
+    status = load_main_exe( NULL, main_argv[1], curdir, 0, &image, module );
+    if (!status)
     {
         char *loader;
 
@@ -1938,18 +2015,14 @@
     if (status)  /* try launching it through start.exe */
     {
         static const char *args[] = { "start.exe", "/exec" };
-        free( nt_name.Buffer );
+        free( image );
         if (*module) NtUnmapViewOfSection( GetCurrentProcess(), *module );
-        load_start_exe( &nt_name, module );
+        load_start_exe( &image, module );
         prepend_argv( args, 2 );
     }
-    else
-    {
-        rebuild_argv();
-        if (NtCurrentTeb64()) NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR] = FALSE;
-    }
-
-    main_wargv = build_wargv( get_dos_path( nt_name.Buffer ));
+    else rebuild_argv();
+
+    main_wargv = build_wargv( get_dos_path( image ));
     cmdline = build_command_line( main_wargv );
 
     TRACE( "image %s cmdline %s dir %s\n",
@@ -1980,7 +2053,7 @@
     put_unicode_string( main_wargv[0], &dst, &params->ImagePathName );
     put_unicode_string( cmdline, &dst, &params->CommandLine );
     put_unicode_string( main_wargv[0], &dst, &params->WindowTitle );
-    free( nt_name.Buffer );
+    free( image );
     free( cmdline );
     free( curdir );
 
@@ -2000,13 +2073,12 @@
  */
 void init_startup_info(void)
 {
-    WCHAR *src, *dst, *env;
+    WCHAR *src, *dst, *env, *image;
     void *module = NULL;
     unsigned int status;
     SIZE_T size, info_size, env_size, env_pos;
     RTL_USER_PROCESS_PARAMETERS *params = NULL;
     struct startup_info_data *info;
-    UNICODE_STRING nt_name;
     USHORT machine;
 
     if (!startup_info_size)
@@ -2080,7 +2152,7 @@
     dst = params->CurrentDirectory.DosPath.Buffer + MAX_PATH;
 
     if (info->dllpath_len) copy_unicode_string( &src, &dst, &params->DllPath, info->dllpath_len );
-    copy_dos_path_string( &src, &dst, &params->ImagePathName, &nt_name, info->imagepath_len );
+    copy_unicode_string( &src, &dst, &params->ImagePathName, info->imagepath_len );
     copy_unicode_string( &src, &dst, &params->CommandLine, info->cmdline_len );
     copy_unicode_string( &src, &dst, &params->WindowTitle, info->title_len );
     copy_unicode_string( &src, &dst, &params->Desktop, info->desktop_len );
@@ -2102,15 +2174,16 @@
     free( env );
     free( info );
 
-    status = load_main_exe( &nt_name, machine, &module );
+    status = load_main_exe( params->ImagePathName.Buffer, NULL, params->CommandLine.Buffer,
+                            machine, &image, &module );
     if (!NT_SUCCESS(status))
     {
-        MESSAGE( "wine: failed to start %s: %x\n", debugstr_us(&params->ImagePathName), status );
+        MESSAGE( "wine: failed to start %s\n", debugstr_us(&params->ImagePathName) );
         NtTerminateProcess( GetCurrentProcess(), status );
     }
     rebuild_argv();
-    main_wargv = build_wargv( params->ImagePathName.Buffer );
-    free( nt_name.Buffer );
+    main_wargv = build_wargv( get_dos_path( image ));
+    free( image );
     init_peb( params, module );
 }
 
@@ -2118,9 +2191,11 @@
 /* helper for create_startup_info */
 static BOOL is_console_handle( HANDLE handle )
 {
+    IO_STATUS_BLOCK io;
     DWORD mode;
 
-    return sync_ioctl( handle, IOCTL_CONDRV_GET_MODE, NULL, 0, &mode, sizeof(mode) ) == STATUS_SUCCESS;
+    return NtDeviceIoControlFile( handle, NULL, NULL, NULL, &io, IOCTL_CONDRV_GET_MODE, NULL, 0,
+                                  &mode, sizeof(mode) ) == STATUS_SUCCESS;
 }
 
 /***********************************************************************
@@ -2131,13 +2206,17 @@
                            const struct pe_image_info *pe_info, DWORD *info_size )
 {
     struct startup_info_data *info;
+    UNICODE_STRING dos_image = *nt_image;
     DWORD size;
     void *ptr;
+
+    dos_image.Buffer = get_dos_path( nt_image->Buffer );
+    dos_image.Length = nt_image->Length - (dos_image.Buffer - nt_image->Buffer) * sizeof(WCHAR);
 
     size = sizeof(*info);
     size += params->CurrentDirectory.DosPath.Length;
     size += params->DllPath.Length;
-    size += nt_image->Length;
+    size += dos_image.Length;
     size += params->CommandLine.Length;
     size += params->WindowTitle.Length;
     size += params->Desktop.Length;
@@ -2176,7 +2255,7 @@
     ptr = info + 1;
     info->curdir_len = append_string( &ptr, params, &params->CurrentDirectory.DosPath );
     info->dllpath_len = append_string( &ptr, params, &params->DllPath );
-    info->imagepath_len = append_string( &ptr, params, nt_image );
+    info->imagepath_len = append_string( &ptr, params, &dos_image );
     info->cmdline_len = append_string( &ptr, params, &params->CommandLine );
     info->title_len = append_string( &ptr, params, &params->WindowTitle );
     info->desktop_len = append_string( &ptr, params, &params->Desktop );
