diff --git a/dlls/ntdll/actctx.c b/dlls/ntdll/actctx.c
--- a/dlls/ntdll/actctx.c
+++ b/dlls/ntdll/actctx.c
@@ -1074,11 +1074,11 @@
     return ret;
 }
 
-static ACTIVATION_CONTEXT *check_actctx( ACTIVATION_CONTEXT *actctx )
-{
-    ACTIVATION_CONTEXT *ret = NULL;
-
-    if (!actctx || actctx == INVALID_HANDLE_VALUE) return NULL;
+static ACTIVATION_CONTEXT *check_actctx( HANDLE h )
+{
+    ACTIVATION_CONTEXT *ret = NULL, *actctx = h;
+
+    if (!h || h == INVALID_HANDLE_VALUE) return NULL;
     __TRY
     {
         if (actctx->magic == ACTCTX_MAGIC) ret = actctx;
@@ -2931,7 +2931,12 @@
     OBJECT_ATTRIBUTES attr;
     IO_STATUS_BLOCK io;
 
-    InitializeObjectAttributes( &attr, name, OBJ_CASE_INSENSITIVE, 0, NULL );
+    attr.Length = sizeof(attr);
+    attr.RootDirectory = 0;
+    attr.Attributes = OBJ_CASE_INSENSITIVE;
+    attr.ObjectName = name;
+    attr.SecurityDescriptor = NULL;
+    attr.SecurityQualityOfService = NULL;
     return NtOpenFile( handle, GENERIC_READ | SYNCHRONIZE, &attr, &io,
                        FILE_SHARE_READ | FILE_SHARE_DELETE, FILE_SYNCHRONOUS_IO_ALERT );
 }
@@ -3028,7 +3033,13 @@
 
     TRACE( "looking for res %s in %s\n", debugstr_w(resname), debugstr_w(filename) );
 
-    InitializeObjectAttributes( &attr, NULL, OBJ_CASE_INSENSITIVE, 0, NULL );
+    attr.Length                   = sizeof(attr);
+    attr.RootDirectory            = 0;
+    attr.ObjectName               = NULL;
+    attr.Attributes               = OBJ_CASE_INSENSITIVE | OBJ_OPENIF;
+    attr.SecurityDescriptor       = NULL;
+    attr.SecurityQualityOfService = NULL;
+
     size.QuadPart = 0;
     status = NtCreateSection( &mapping, STANDARD_RIGHTS_REQUIRED | SECTION_QUERY | SECTION_MAP_READ,
                               &attr, &size, PAGE_READONLY, SEC_COMMIT, file );
@@ -3068,7 +3079,13 @@
 
     TRACE( "loading manifest file %s\n", debugstr_w(filename) );
 
-    InitializeObjectAttributes( &attr, NULL, OBJ_CASE_INSENSITIVE, 0, NULL );
+    attr.Length                   = sizeof(attr);
+    attr.RootDirectory            = 0;
+    attr.ObjectName               = NULL;
+    attr.Attributes               = OBJ_CASE_INSENSITIVE | OBJ_OPENIF;
+    attr.SecurityDescriptor       = NULL;
+    attr.SecurityQualityOfService = NULL;
+
     size.QuadPart = 0;
     status = NtCreateSection( &mapping, STANDARD_RIGHTS_REQUIRED | SECTION_QUERY | SECTION_MAP_READ,
                               &attr, &size, PAGE_READONLY, SEC_COMMIT, file );
@@ -3139,64 +3156,24 @@
     return status;
 }
 
-static void append_field( WCHAR *buffer, const WCHAR *str, unsigned int maxlen )
-{
-    static const WCHAR valid_chars[] = L"-.0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";
-    unsigned int len;
-
-    if (!*str) return;
-    buffer += wcslen( buffer );
-    for (len = 0; *str; str++) if (wcschr( valid_chars, *str )) buffer[len++] = *str;
-    if (len > maxlen)
-    {
-        unsigned int pos = maxlen / 2;
-        buffer[pos - 1] = buffer[pos] = '.';
-        memmove( buffer + pos + 1, buffer + len - pos + 1, (pos - 1) * sizeof(WCHAR) );
-        len = maxlen;
-    }
-    buffer[len++] = '_';
-    buffer[len] = 0;
-}
-
-static WCHAR *build_manifest_filter( struct assembly_identity *ai )
-{
-    WCHAR version[32];
-    unsigned int buflen = 20;
-    WCHAR *ret;
-
-    swprintf( version, ARRAY_SIZE(version), L"_%u.%u.*.*_", ai->version.major, ai->version.minor );
-
-    buflen += wcslen( ai->arch ) + 1;
-    buflen += wcslen( ai->name ) + 1;
-    buflen += wcslen( ai->public_key ) + 1;
-    buflen += wcslen( version ) + 1;
-    if (ai->language) buflen += wcslen( ai->language );
-    if (!(ret = RtlAllocateHeap( GetProcessHeap(), 0, buflen * sizeof(WCHAR) ))) return NULL;
-    ret[0] = 0;
-    append_field( ret, ai->arch, 16 );
-    append_field( ret, ai->name, 40 );
-    wcscat( ret, ai->public_key );
-    wcscat( ret, version );
-    if (!ai->language || !wcsicmp( ai->language, L"*" ) || !wcsicmp( ai->language, L"neutral" ))
-        wcscat( ret, L"*_" );
-    else
-        append_field( ret, ai->language, 8 );
-
-    wcscat( ret, L"*.manifest" );
-    return ret;
-}
-
 static WCHAR *lookup_manifest_file( HANDLE dir, struct assembly_identity *ai )
 {
+    static const WCHAR lookup_fmtW[] = L"%s_%s_%s_%u.%u.*.*_%s_*.manifest";
     static const WCHAR wine_trailerW[] = {'d','e','a','d','b','e','e','f','.','m','a','n','i','f','e','s','t'};
 
     WCHAR *lookup, *ret = NULL;
     UNICODE_STRING lookup_us;
     IO_STATUS_BLOCK io;
-    unsigned int data_pos = 0, data_len;
+    const WCHAR *lang = ai->language;
+    unsigned int data_pos = 0, data_len, len;
     char buffer[8192];
 
-    if (!(lookup = build_manifest_filter( ai ))) return NULL;
+    if (!lang || !wcsicmp( lang, L"neutral" )) lang = L"*";
+
+    len = wcslen(ai->arch) + wcslen(ai->name) + wcslen(ai->public_key) + wcslen(lang) + 20 + ARRAY_SIZE(lookup_fmtW);
+    if (!(lookup = RtlAllocateHeap( GetProcessHeap(), 0, len * sizeof(WCHAR) ))) return NULL;
+    swprintf( lookup, len, lookup_fmtW, ai->arch, ai->name, ai->public_key,
+              ai->version.major, ai->version.minor, lang );
     RtlInitUnicodeString( &lookup_us, lookup );
 
 #ifdef __arm64ec__
@@ -3287,7 +3264,14 @@
         return STATUS_NO_SUCH_FILE;
     }
     RtlFreeHeap( GetProcessHeap(), 0, path );
-    InitializeObjectAttributes( &attr, &path_us, OBJ_CASE_INSENSITIVE, 0, NULL );
+
+    attr.Length = sizeof(attr);
+    attr.RootDirectory = 0;
+    attr.Attributes = OBJ_CASE_INSENSITIVE;
+    attr.ObjectName = &path_us;
+    attr.SecurityDescriptor = NULL;
+    attr.SecurityQualityOfService = NULL;
+
     if (!NtOpenFile( &handle, GENERIC_READ | SYNCHRONIZE, &attr, &io, FILE_SHARE_READ | FILE_SHARE_WRITE,
                      FILE_DIRECTORY_FILE | FILE_SYNCHRONOUS_IO_NONALERT ))
     {
@@ -3327,69 +3311,30 @@
     return io.Status;
 }
 
-static NTSTATUS open_manifest_file( struct actctx_loader *acl, struct assembly_identity *ai,
-                                    const WCHAR *lang, const WCHAR *directory, WCHAR *buffer, DWORD len )
-{
-    UNICODE_STRING nameW;
-    NTSTATUS status;
-    HANDLE file;
-    WCHAR *p = buffer + wcslen(buffer);
-
-    nameW.Buffer = NULL;
-    if (*lang) p += swprintf( p, len - (p - buffer), L"%s\\", lang );
-
-    swprintf( p, len - (p - buffer), L"%s.dll", ai->name );
-    if (RtlDosPathNameToNtPathName_U( buffer, &nameW, NULL, NULL ))
-    {
-        status = open_nt_file( &file, &nameW );
-        if (!status)
-        {
-            status = get_manifest_in_pe_file( acl, ai, nameW.Buffer, directory, FALSE, file, NULL, 0 );
-            NtClose( file );
-            if (status == STATUS_SUCCESS) goto done;
-        }
-        RtlFreeUnicodeString( &nameW );
-    }
-
-    swprintf( p, len - (p - buffer), L"%s.manifest", ai->name );
-    if (RtlDosPathNameToNtPathName_U( buffer, &nameW, NULL, NULL ))
-    {
-        status = open_nt_file( &file, &nameW );
-        if (!status)
-        {
-            status = get_manifest_in_manifest_file( acl, ai, nameW.Buffer, directory, FALSE, file );
-            NtClose( file );
-            goto done;
-        }
-        RtlFreeUnicodeString( &nameW );
-    }
-    status = STATUS_SXS_ASSEMBLY_NOT_FOUND;
-done:
-    RtlFreeUnicodeString( &nameW );
-    return status;
-}
-
 static NTSTATUS lookup_assembly(struct actctx_loader* acl,
                                 struct assembly_identity* ai)
 {
+    unsigned int i;
     WCHAR *buffer, *p, *directory;
-    const WCHAR *lang = ai->language;
     NTSTATUS status;
-    DWORD len, total;
-
-    TRACE( "looking for name=%s version=%s arch=%s lang=%s\n",
-           debugstr_w(ai->name), debugstr_version(&ai->version),
-           debugstr_w(ai->arch), debugstr_w(ai->language) );
+    UNICODE_STRING nameW;
+    HANDLE file;
+    DWORD len;
+
+    TRACE( "looking for name=%s version=%s arch=%s\n",
+           debugstr_w(ai->name), debugstr_version(&ai->version), debugstr_w(ai->arch) );
 
     if ((status = lookup_winsxs(acl, ai)) != STATUS_NO_SUCH_FILE) return status;
 
-    if (!lang || !wcsicmp( lang, L"neutral" ) || !wcscmp( lang, L"*")) lang = L"";
+    /* FIXME: add support for language specific lookup */
 
     len = max(RtlGetFullPathName_U(acl->actctx->assemblies->manifest.info, 0, NULL, NULL) / sizeof(WCHAR),
         wcslen(acl->actctx->appdir.info));
-    total = len + 2 * wcslen(ai->name) + wcslen(lang) + 12;
-
-    if (!(buffer = RtlAllocateHeap( GetProcessHeap(), 0, total * sizeof(WCHAR) ))) return STATUS_NO_MEMORY;
+
+    nameW.Buffer = NULL;
+    if (!(buffer = RtlAllocateHeap( GetProcessHeap(), 0,
+                                    (len + 2 * wcslen(ai->name) + 2) * sizeof(WCHAR) + sizeof(L".manifest") )))
+        return STATUS_NO_MEMORY;
 
     if (!(directory = build_assembly_dir( ai )))
     {
@@ -3407,25 +3352,47 @@
      */
     wcscpy( buffer, acl->actctx->appdir.info );
     p = buffer + wcslen(buffer);
-
-    status = open_manifest_file( acl, ai, lang, directory, buffer, total );
-    if (status != STATUS_SXS_ASSEMBLY_NOT_FOUND) goto done;
-
-    swprintf( buffer, total, L"%s%s\\", acl->actctx->appdir.info, ai->name );
-    status = open_manifest_file( acl, ai, lang, directory, buffer, total );
-    if (status != STATUS_SXS_ASSEMBLY_NOT_FOUND) goto done;
-
-    if (RtlGetFullPathName_U( acl->actctx->assemblies->manifest.info, len * sizeof(WCHAR), buffer, &p ))
-    {
-        *p = 0;
-        status = open_manifest_file( acl, ai, lang, directory, buffer, total );
-        if (status != STATUS_SXS_ASSEMBLY_NOT_FOUND) goto done;
-
-        swprintf( p, total - (p - buffer), L"%s\\", ai->name );
-        status = open_manifest_file( acl, ai, lang, directory, buffer, total );
-    }
-
-done:
+    for (i = 0; i < 4; i++)
+    {
+        if (i == 2)
+        {
+            struct assembly *assembly = acl->actctx->assemblies;
+            if (!RtlGetFullPathName_U(assembly->manifest.info, len * sizeof(WCHAR), buffer, &p)) break;
+        }
+        else *p++ = '\\';
+
+        wcscpy( p, ai->name );
+        p += wcslen(p);
+
+        wcscpy( p, L".dll" );
+        if (RtlDosPathNameToNtPathName_U( buffer, &nameW, NULL, NULL ))
+        {
+            status = open_nt_file( &file, &nameW );
+            if (!status)
+            {
+                status = get_manifest_in_pe_file( acl, ai, nameW.Buffer, directory, FALSE, file, NULL, 0 );
+                NtClose( file );
+                if (status == STATUS_SUCCESS)
+                    break;
+            }
+            RtlFreeUnicodeString( &nameW );
+        }
+
+        wcscpy( p, L".manifest" );
+        if (RtlDosPathNameToNtPathName_U( buffer, &nameW, NULL, NULL ))
+        {
+            status = open_nt_file( &file, &nameW );
+            if (!status)
+            {
+                status = get_manifest_in_manifest_file( acl, ai, nameW.Buffer, directory, FALSE, file );
+                NtClose( file );
+                break;
+            }
+            RtlFreeUnicodeString( &nameW );
+        }
+        status = STATUS_SXS_ASSEMBLY_NOT_FOUND;
+    }
+    RtlFreeUnicodeString( &nameW );
     RtlFreeHeap( GetProcessHeap(), 0, directory );
     RtlFreeHeap( GetProcessHeap(), 0, buffer );
     return status;
@@ -3454,7 +3421,7 @@
     return status;
 }
 
-static ACTIVATION_CONTEXT * get_current_actctx_no_addref(void)
+static HANDLE get_current_actctx_no_addref(void)
 {
     ACTIVATION_CONTEXT_STACK *actctx_stack = NtCurrentTeb()->ActivationContextStackPointer;
 
@@ -3465,36 +3432,36 @@
 }
 
 /* find the appropriate activation context for RtlQueryInformationActivationContext */
-static NTSTATUS find_query_actctx( ACTIVATION_CONTEXT **actctx, DWORD flags, ULONG class )
+static NTSTATUS find_query_actctx( HANDLE *handle, DWORD flags, ULONG class )
 {
     NTSTATUS status = STATUS_SUCCESS;
 
     if (flags & QUERY_ACTCTX_FLAG_USE_ACTIVE_ACTCTX)
     {
-        if (*actctx) return STATUS_INVALID_PARAMETER;
-
-        *actctx = get_current_actctx_no_addref();
+        if (*handle) return STATUS_INVALID_PARAMETER;
+
+        *handle = get_current_actctx_no_addref();
     }
     else if (flags & (QUERY_ACTCTX_FLAG_ACTCTX_IS_ADDRESS|QUERY_ACTCTX_FLAG_ACTCTX_IS_HMODULE))
     {
         ULONG_PTR magic;
         LDR_DATA_TABLE_ENTRY *pldr;
 
-        if (!*actctx) return STATUS_INVALID_PARAMETER;
+        if (!*handle) return STATUS_INVALID_PARAMETER;
 
         LdrLockLoaderLock( 0, NULL, &magic );
-        if (!LdrFindEntryForAddress( *actctx, &pldr ))
-        {
-            if ((flags & QUERY_ACTCTX_FLAG_ACTCTX_IS_HMODULE) && *actctx != pldr->DllBase)
+        if (!LdrFindEntryForAddress( *handle, &pldr ))
+        {
+            if ((flags & QUERY_ACTCTX_FLAG_ACTCTX_IS_HMODULE) && *handle != pldr->DllBase)
                 status = STATUS_DLL_NOT_FOUND;
             else
-                *actctx = pldr->ActivationContext;
+                *handle = pldr->ActivationContext;
         }
         else status = STATUS_DLL_NOT_FOUND;
         LdrUnlockLoaderLock( 0, magic );
     }
-    else if (!*actctx && (class != ActivationContextBasicInformation))
-        *actctx = process_actctx;
+    else if (!*handle && (class != ActivationContextBasicInformation))
+        *handle = process_actctx;
 
     return status;
 }
@@ -5267,7 +5234,7 @@
 void actctx_init(void)
 {
     ACTCTXW ctx;
-    ACTIVATION_CONTEXT *actctx;
+    HANDLE handle;
 
     ctx.cbSize   = sizeof(ctx);
     ctx.lpSource = NULL;
@@ -5275,7 +5242,7 @@
     ctx.hModule  = NtCurrentTeb()->Peb->ImageBaseAddress;
     ctx.lpResourceName = (LPCWSTR)CREATEPROCESS_MANIFEST_RESOURCE_ID;
 
-    if (!RtlCreateActivationContext( &actctx, &ctx )) process_actctx = check_actctx( actctx );
+    if (!RtlCreateActivationContext( &handle, &ctx )) process_actctx = check_actctx(handle);
 
     NtCurrentTeb()->Peb->ActivationContextData = process_actctx;
 }
@@ -5288,7 +5255,7 @@
  *
  * FIXME: function signature/prototype is wrong
  */
-NTSTATUS WINAPI RtlCreateActivationContext( ACTIVATION_CONTEXT **new_actctx, const void *ptr )
+NTSTATUS WINAPI RtlCreateActivationContext( HANDLE *handle, const void *ptr )
 {
     const ACTCTXW *pActCtx = ptr;  /* FIXME: not the right structure */
     const WCHAR *directory = NULL;
@@ -5424,7 +5391,7 @@
     if (status == STATUS_SUCCESS) status = parse_depend_manifests(&acl);
     free_depend_manifests( &acl );
 
-    if (status == STATUS_SUCCESS) *new_actctx = actctx;
+    if (status == STATUS_SUCCESS) *handle = actctx;
     else actctx_release( actctx );
     return status;
 
@@ -5438,18 +5405,22 @@
 /***********************************************************************
  *		RtlAddRefActivationContext (NTDLL.@)
  */
-void WINAPI RtlAddRefActivationContext( ACTIVATION_CONTEXT *actctx )
-{
-    if (check_actctx( actctx )) actctx_addref( actctx );
+void WINAPI RtlAddRefActivationContext( HANDLE handle )
+{
+    ACTIVATION_CONTEXT *actctx;
+
+    if ((actctx = check_actctx( handle ))) actctx_addref( actctx );
 }
 
 
 /******************************************************************
  *		RtlReleaseActivationContext (NTDLL.@)
  */
-void WINAPI RtlReleaseActivationContext( ACTIVATION_CONTEXT *actctx )
-{
-    if (check_actctx( actctx )) actctx_release( actctx );
+void WINAPI RtlReleaseActivationContext( HANDLE handle )
+{
+    ACTIVATION_CONTEXT *actctx;
+
+    if ((actctx = check_actctx( handle ))) actctx_release( actctx );
 }
 
 /******************************************************************
@@ -5457,76 +5428,38 @@
  *
  * FIXME: function prototype might be wrong
  */
-NTSTATUS WINAPI RtlZombifyActivationContext( ACTIVATION_CONTEXT *actctx )
-{
-    FIXME("%p: stub\n", actctx);
+NTSTATUS WINAPI RtlZombifyActivationContext( HANDLE handle )
+{
+    FIXME("%p: stub\n", handle);
     return STATUS_NOT_IMPLEMENTED;
 }
 
 /******************************************************************
  *		RtlActivateActivationContext (NTDLL.@)
  */
-NTSTATUS WINAPI RtlActivateActivationContext( ULONG unknown, ACTIVATION_CONTEXT *actctx, PULONG_PTR cookie )
-{
-    return RtlActivateActivationContextEx( 0, NtCurrentTeb(), actctx, cookie );
-}
-
-/******************************************************************
- *              RtlActivateActivationContextUnsafeFast (NTDLL.@)
- *
- * FIXME: function prototype might be wrong
- */
-DEFINE_FASTCALL_WRAPPER( RtlActivateActivationContextUnsafeFast, 8 )
-RTL_ACTIVATION_CONTEXT_STACK_FRAME * FASTCALL RtlActivateActivationContextUnsafeFast( RTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_EXTENDED *frame_extended,
-                                                                                      ACTIVATION_CONTEXT *actctx )
-{
-    ACTIVATION_CONTEXT_STACK *actctx_stack = NtCurrentTeb()->ActivationContextStackPointer;
-
-    TRACE( "%p %p\n", frame_extended, actctx );
-
-    frame_extended->Frame.Previous = actctx_stack->ActiveFrame;
-    frame_extended->Frame.ActivationContext = actctx;
-    frame_extended->Frame.Flags = 0;
-    actctx_stack->ActiveFrame = &frame_extended->Frame;
-    RtlAddRefActivationContext( actctx );
-    return &frame_extended->Frame;
-}
-
-/******************************************************************
- *              RtlDeactivateActivationContextUnsafeFast (NTDLL.@)
- *
- * FIXME: function prototype might be wrong
- */
-DEFINE_FASTCALL1_WRAPPER( RtlDeactivateActivationContextUnsafeFast )
-void FASTCALL RtlDeactivateActivationContextUnsafeFast( RTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_EXTENDED *frame_extended )
-{
-    ACTIVATION_CONTEXT_STACK *actctx_stack = NtCurrentTeb()->ActivationContextStackPointer;
-    ACTIVATION_CONTEXT *actctx;
-
-    TRACE( "%p\n", frame_extended );
-
-    actctx = actctx_stack->ActiveFrame->ActivationContext;
-    actctx_stack->ActiveFrame = frame_extended->Frame.Previous;
-    RtlReleaseActivationContext( actctx );
-}
+NTSTATUS WINAPI RtlActivateActivationContext( ULONG unknown, HANDLE handle, PULONG_PTR cookie )
+{
+    return RtlActivateActivationContextEx( 0, NtCurrentTeb(), handle, cookie );
+}
+
 
 /******************************************************************
  *		RtlActivateActivationContextEx (NTDLL.@)
  */
-NTSTATUS WINAPI RtlActivateActivationContextEx( ULONG flags, TEB *teb, ACTIVATION_CONTEXT *actctx, ULONG_PTR *cookie )
+NTSTATUS WINAPI RtlActivateActivationContextEx( ULONG flags, TEB *teb, HANDLE handle, ULONG_PTR *cookie )
 {
     ACTIVATION_CONTEXT_STACK *actctx_stack = teb->ActivationContextStackPointer;
     RTL_ACTIVATION_CONTEXT_STACK_FRAME *frame;
 
     frame = RtlAllocateHeap( GetProcessHeap(), HEAP_GENERATE_EXCEPTIONS, sizeof(*frame) );
     frame->Previous = actctx_stack->ActiveFrame;
-    frame->ActivationContext = actctx;
-    frame->Flags = NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED;
+    frame->ActivationContext = handle;
+    frame->Flags = 0;
     actctx_stack->ActiveFrame = frame;
-    RtlAddRefActivationContext( actctx );
+    RtlAddRefActivationContext( handle );
 
     *cookie = (ULONG_PTR)frame;
-    TRACE( "%p cookie=%Ix\n", actctx, *cookie );
+    TRACE( "%p cookie=%Ix\n", handle, *cookie );
     return STATUS_SUCCESS;
 }
 
@@ -5586,8 +5519,7 @@
     {
         RTL_ACTIVATION_CONTEXT_STACK_FRAME *prev = frame->Previous;
         RtlReleaseActivationContext( frame->ActivationContext );
-        if (frame->Flags & NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED)
-            RtlFreeHeap( GetProcessHeap(), 0, frame );
+        RtlFreeHeap( GetProcessHeap(), 0, frame );
         frame = prev;
     }
     actctx_stack->ActiveFrame = NULL;
@@ -5597,9 +5529,9 @@
 /******************************************************************
  *		RtlGetActiveActivationContext (NTDLL.@)
  */
-NTSTATUS WINAPI RtlGetActiveActivationContext( ACTIVATION_CONTEXT **actctx )
-{
-    RtlAddRefActivationContext( *actctx = get_current_actctx_no_addref() );
+NTSTATUS WINAPI RtlGetActiveActivationContext( HANDLE *handle )
+{
+    RtlAddRefActivationContext( *handle = get_current_actctx_no_addref() );
     return STATUS_SUCCESS;
 }
 
@@ -5607,13 +5539,13 @@
 /******************************************************************
  *		RtlIsActivationContextActive (NTDLL.@)
  */
-BOOLEAN WINAPI RtlIsActivationContextActive( ACTIVATION_CONTEXT *actctx )
+BOOLEAN WINAPI RtlIsActivationContextActive( HANDLE handle )
 {
     ACTIVATION_CONTEXT_STACK *actctx_stack = NtCurrentTeb()->ActivationContextStackPointer;
     RTL_ACTIVATION_CONTEXT_STACK_FRAME *frame;
 
     for (frame = actctx_stack->ActiveFrame; frame; frame = frame->Previous)
-        if (frame->ActivationContext == actctx) return TRUE;
+        if (frame->ActivationContext == handle) return TRUE;
     return FALSE;
 }
 
@@ -5624,17 +5556,18 @@
  * Get information about an activation context.
  * FIXME: function signature/prototype may be wrong
  */
-NTSTATUS WINAPI RtlQueryInformationActivationContext( ULONG flags, ACTIVATION_CONTEXT *actctx, PVOID subinst,
+NTSTATUS WINAPI RtlQueryInformationActivationContext( ULONG flags, HANDLE handle, PVOID subinst,
                                                       ULONG class, PVOID buffer,
                                                       SIZE_T bufsize, SIZE_T *retlen )
 {
+    ACTIVATION_CONTEXT *actctx;
     NTSTATUS status;
 
-    TRACE("%08lx %p %p %lu %p %Id %p\n", flags, actctx,
+    TRACE("%08lx %p %p %lu %p %Id %p\n", flags, handle,
           subinst, class, buffer, bufsize, retlen);
 
     if (retlen) *retlen = 0;
-    if ((status = find_query_actctx( &actctx, flags, class ))) return status;
+    if ((status = find_query_actctx( &handle, flags, class ))) return status;
 
     switch (class)
     {
@@ -5645,9 +5578,9 @@
             if (retlen) *retlen = sizeof(*info);
             if (!info || bufsize < sizeof(*info)) return STATUS_BUFFER_TOO_SMALL;
 
-            info->hActCtx = actctx;
+            info->hActCtx = handle;
             info->dwFlags = 0;  /* FIXME */
-            if (!(flags & QUERY_ACTCTX_FLAG_NO_ADDREF)) RtlAddRefActivationContext( actctx );
+            if (!(flags & QUERY_ACTCTX_FLAG_NO_ADDREF)) RtlAddRefActivationContext( handle );
         }
         break;
 
@@ -5658,7 +5591,7 @@
             SIZE_T len, manifest_len = 0, config_len = 0, appdir_len = 0;
             LPWSTR ptr;
 
-            if (!check_actctx(actctx)) return STATUS_INVALID_PARAMETER;
+            if (!(actctx = check_actctx(handle))) return STATUS_INVALID_PARAMETER;
 
             if (actctx->num_assemblies) assembly = actctx->assemblies;
 
@@ -5713,7 +5646,7 @@
             SIZE_T len, id_len = 0, ad_len = 0, path_len = 0;
             LPWSTR ptr;
 
-            if (!check_actctx(actctx)) return STATUS_INVALID_PARAMETER;
+            if (!(actctx = check_actctx(handle))) return STATUS_INVALID_PARAMETER;
             if (!subinst) return STATUS_INVALID_PARAMETER;
 
             index = *(DWORD*)subinst;
@@ -5782,7 +5715,7 @@
             SIZE_T len, dll_len = 0;
             LPWSTR ptr;
 
-            if (!check_actctx(actctx)) return STATUS_INVALID_PARAMETER;
+            if (!(actctx = check_actctx(handle))) return STATUS_INVALID_PARAMETER;
             if (!acqi) return STATUS_INVALID_PARAMETER;
 
             if (acqi->ulAssemblyIndex >= actctx->num_assemblies)
@@ -5826,7 +5759,7 @@
             ULONG num_compat_contexts = 0, n;
             SIZE_T len;
 
-            if (!check_actctx(actctx)) return STATUS_INVALID_PARAMETER;
+            if (!(actctx = check_actctx(handle))) return STATUS_INVALID_PARAMETER;
 
             if (actctx->num_assemblies) assembly = actctx->assemblies;
 
@@ -5851,7 +5784,7 @@
             struct assembly *assembly;
             SIZE_T len;
 
-            if (!check_actctx(actctx)) return STATUS_INVALID_PARAMETER;
+            if (!(actctx = check_actctx(handle))) return STATUS_INVALID_PARAMETER;
 
             len = sizeof(*acrli);
             if (retlen) *retlen = len;
@@ -5871,16 +5804,6 @@
         return STATUS_NOT_IMPLEMENTED;
     }
     return STATUS_SUCCESS;
-}
-
-/******************************************************************
- *              RtlQueryInformationActiveActivationContext (NTDLL.@)
- */
-NTSTATUS WINAPI RtlQueryInformationActiveActivationContext( ULONG class, PVOID buffer,
-                                                            SIZE_T buffer_size, SIZE_T *return_length )
-{
-    return RtlQueryInformationActivationContext( QUERY_ACTCTX_FLAG_USE_ACTIVE_ACTCTX, NULL, NULL,
-                                                 class, buffer, buffer_size, return_length );
 }
 
 /***********************************************************************
@@ -5968,10 +5891,11 @@
 /***********************************************************************
  *		RtlQueryActivationContextApplicationSettings (NTDLL.@)
  */
-NTSTATUS WINAPI RtlQueryActivationContextApplicationSettings( DWORD flags, ACTIVATION_CONTEXT *actctx, const WCHAR *ns,
+NTSTATUS WINAPI RtlQueryActivationContextApplicationSettings( DWORD flags, HANDLE handle, const WCHAR *ns,
                                                               const WCHAR *settings, WCHAR *buffer,
                                                               SIZE_T size, SIZE_T *written )
 {
+    ACTIVATION_CONTEXT *actctx;
     const WCHAR *res;
 
     if (flags)
@@ -5992,8 +5916,8 @@
     }
     else ns = windowsSettings2005NSW;
 
-    if (!actctx) actctx = process_actctx;
-    if (!check_actctx( actctx )) return STATUS_INVALID_PARAMETER;
+    if (!handle) handle = process_actctx;
+    if (!(actctx = check_actctx( handle ))) return STATUS_INVALID_PARAMETER;
 
     if (!(res = find_app_settings( actctx, settings, ns ))) return STATUS_SXS_KEY_NOT_FOUND;
 
