diff --git a/dlls/winemac.drv/window.c b/dlls/winemac.drv/window.c
--- a/dlls/winemac.drv/window.c
+++ b/dlls/winemac.drv/window.c
@@ -177,6 +177,7 @@
     if ((data = calloc(1, sizeof(*data))))
     {
         data->hwnd = hwnd;
+        data->swap_interval = 1;
         pthread_mutex_lock(&win_data_mutex);
         if (!win_datas)
             win_datas = CFDictionaryCreateMutable(NULL, 0, NULL, NULL);
@@ -226,6 +227,36 @@
     macdrv_window ret = NULL;
     if (data && (data->on_screen || !require_on_screen))
         ret = data->cocoa_window;
+    release_win_data(data);
+    return ret;
+}
+
+
+/***********************************************************************
+ *              macdrv_get_cocoa_view
+ *
+ * Return the Cocoa view associated with a window
+ */
+macdrv_view macdrv_get_cocoa_view(HWND hwnd)
+{
+    struct macdrv_win_data *data = get_win_data(hwnd);
+    macdrv_view ret = data ? data->cocoa_view : NULL;
+
+    release_win_data(data);
+    return ret;
+}
+
+
+/***********************************************************************
+ *              macdrv_get_client_cocoa_view
+ *
+ * Return the Cocoa view associated with a window's client area
+ */
+macdrv_view macdrv_get_client_cocoa_view(HWND hwnd)
+{
+    struct macdrv_win_data *data = get_win_data(hwnd);
+    macdrv_view ret = data ? data->client_cocoa_view : NULL;
+
     release_win_data(data);
     return ret;
 }
@@ -445,6 +476,27 @@
 
 
 /**********************************************************************
+ *              create_client_cocoa_view
+ *
+ * Create the Cocoa view for a window's client area
+ */
+static void create_client_cocoa_view(struct macdrv_win_data *data)
+{
+    RECT rect = data->rects.client;
+    OffsetRect(&rect, -data->rects.visible.left, -data->rects.visible.top);
+
+    if (data->client_cocoa_view)
+        macdrv_set_view_frame(data->client_cocoa_view, cgrect_from_rect(rect));
+    else
+    {
+        data->client_cocoa_view = macdrv_create_view(cgrect_from_rect(rect));
+        macdrv_set_view_hidden(data->client_cocoa_view, FALSE);
+    }
+    macdrv_set_view_superview(data->client_cocoa_view, data->cocoa_view, data->cocoa_window, NULL, NULL);
+}
+
+
+/**********************************************************************
  *              create_cocoa_window
  *
  * Create the whole Mac window for a given window
@@ -484,6 +536,7 @@
 
     data->cocoa_window = macdrv_create_cocoa_window(&wf, frame, data->hwnd, thread_data->queue);
     if (!data->cocoa_window) goto done;
+    create_client_cocoa_view(data);
 
     set_cocoa_window_properties(data);
 
@@ -520,6 +573,71 @@
 }
 
 
+/**********************************************************************
+ *              create_cocoa_view
+ *
+ * Create the Cocoa view for a given Windows child window
+ */
+static void create_cocoa_view(struct macdrv_win_data *data)
+{
+    BOOL equal = EqualRect(&data->rects.window, &data->rects.client);
+    CGRect frame = cgrect_from_rect(data->rects.window);
+
+    data->shaped = FALSE;
+    data->rects.visible = data->rects.window;
+
+    TRACE("creating %p window %s whole %s client %s\n", data->hwnd, wine_dbgstr_rect(&data->rects.window),
+          wine_dbgstr_rect(&data->rects.visible), wine_dbgstr_rect(&data->rects.client));
+
+    if (!equal)
+        data->cocoa_view = macdrv_create_view(frame);
+    create_client_cocoa_view(data);
+    if (equal)
+    {
+        data->cocoa_view = data->client_cocoa_view;
+        macdrv_set_view_hidden(data->cocoa_view, TRUE);
+        macdrv_set_view_frame(data->cocoa_view, frame);
+    }
+}
+
+
+/**********************************************************************
+ *              destroy_cocoa_view
+ *
+ * Destroy the Cocoa view for a given window.
+ */
+static void destroy_cocoa_view(struct macdrv_win_data *data)
+{
+    if (!data->cocoa_view) return;
+
+    TRACE("win %p Cocoa view %p\n", data->hwnd, data->cocoa_view);
+
+    if (data->cocoa_view != data->client_cocoa_view)
+        macdrv_dispose_view(data->cocoa_view);
+    data->cocoa_view = NULL;
+    data->on_screen = FALSE;
+}
+
+
+/***********************************************************************
+ *              set_cocoa_view_parent
+ */
+static void set_cocoa_view_parent(struct macdrv_win_data *data, HWND parent)
+{
+    struct macdrv_win_data *parent_data = get_win_data(parent);
+    macdrv_window cocoa_window = parent_data ? parent_data->cocoa_window : NULL;
+    macdrv_view superview = parent_data ? parent_data->client_cocoa_view : NULL;
+
+    TRACE("win %p/%p parent %p/%p\n", data->hwnd, data->cocoa_view, parent, cocoa_window ? (void*)cocoa_window : (void*)superview);
+
+    if (!cocoa_window && !superview)
+        WARN("hwnd %p new parent %p has no Cocoa window or view in this process\n", data->hwnd, parent);
+
+    macdrv_set_view_superview(data->cocoa_view, superview, cocoa_window, NULL, NULL);
+    release_win_data(parent_data);
+}
+
+
 /***********************************************************************
  *              macdrv_create_win_data
  *
@@ -537,6 +655,9 @@
         macdrv_init_thread_data();
         return NULL;
     }
+
+    /* don't create win data for HWND_MESSAGE windows */
+    if (parent != NtUserGetDesktopWindow() && !NtUserGetAncestor(parent, GA_PARENT)) return NULL;
 
     if (!(data = alloc_win_data(hwnd))) return NULL;
     data->rects = *rects;
@@ -548,10 +669,16 @@
                hwnd, data->cocoa_window, wine_dbgstr_rect(&data->rects.window),
                wine_dbgstr_rect(&data->rects.visible), wine_dbgstr_rect(&data->rects.client));
     }
-
-    TRACE("win %p/%p window %s whole %s client %s\n",
-           hwnd, data->cocoa_window, wine_dbgstr_rect(&data->rects.window),
-           wine_dbgstr_rect(&data->rects.visible), wine_dbgstr_rect(&data->rects.client));
+    else
+    {
+        create_cocoa_view(data);
+        TRACE("win %p/%p window %s whole %s client %s\n",
+               hwnd, data->cocoa_view, wine_dbgstr_rect(&data->rects.window),
+               wine_dbgstr_rect(&data->rects.visible), wine_dbgstr_rect(&data->rects.client));
+
+        set_cocoa_view_parent(data, parent);
+    }
+
     return data;
 }
 
@@ -627,41 +754,51 @@
  */
 static void show_window(struct macdrv_win_data *data)
 {
-    HWND prev = NULL;
-    HWND next = NULL;
-    macdrv_window prev_window = NULL;
-    macdrv_window next_window = NULL;
-    BOOL activate = FALSE;
-    GUITHREADINFO info;
-
-    /* find window that this one must be after */
-    prev = NtUserGetWindowRelative(data->hwnd, GW_HWNDPREV);
-    while (prev && !((NtUserGetWindowLongW(prev, GWL_STYLE) & (WS_VISIBLE | WS_MINIMIZE)) == WS_VISIBLE &&
-                     (prev_window = macdrv_get_cocoa_window(prev, TRUE))))
-        prev = NtUserGetWindowRelative(prev, GW_HWNDPREV);
-    if (!prev_window)
-    {
-        /* find window that this one must be before */
-        next = NtUserGetWindowRelative(data->hwnd, GW_HWNDNEXT);
-        while (next && !((NtUserGetWindowLongW(next, GWL_STYLE) & (WS_VISIBLE | WS_MINIMIZE)) == WS_VISIBLE &&
-                         (next_window = macdrv_get_cocoa_window(next, TRUE))))
-            next = NtUserGetWindowRelative(next, GW_HWNDNEXT);
-    }
-
-    TRACE("win %p/%p below %p/%p above %p/%p\n",
-          data->hwnd, data->cocoa_window, prev, prev_window, next, next_window);
-
-    if (!prev_window)
-        activate = activate_on_focus_time && (NtGetTickCount() - activate_on_focus_time < 2000);
-    macdrv_order_cocoa_window(data->cocoa_window, prev_window, next_window, activate);
-    data->on_screen = TRUE;
-
-    info.cbSize = sizeof(info);
-    if (NtUserGetGUIThreadInfo(NtUserGetWindowThread(data->hwnd, NULL), &info) && info.hwndFocus &&
-        (data->hwnd == info.hwndFocus || NtUserIsChild(data->hwnd, info.hwndFocus)))
-        set_focus(info.hwndFocus, FALSE);
-    if (activate)
-        activate_on_focus_time = 0;
+    if (data->cocoa_window)
+    {
+        HWND prev = NULL;
+        HWND next = NULL;
+        macdrv_window prev_window = NULL;
+        macdrv_window next_window = NULL;
+        BOOL activate = FALSE;
+        GUITHREADINFO info;
+
+        /* find window that this one must be after */
+        prev = NtUserGetWindowRelative(data->hwnd, GW_HWNDPREV);
+        while (prev && !((NtUserGetWindowLongW(prev, GWL_STYLE) & (WS_VISIBLE | WS_MINIMIZE)) == WS_VISIBLE &&
+                         (prev_window = macdrv_get_cocoa_window(prev, TRUE))))
+            prev = NtUserGetWindowRelative(prev, GW_HWNDPREV);
+        if (!prev_window)
+        {
+            /* find window that this one must be before */
+            next = NtUserGetWindowRelative(data->hwnd, GW_HWNDNEXT);
+            while (next && !((NtUserGetWindowLongW(next, GWL_STYLE) & (WS_VISIBLE | WS_MINIMIZE)) == WS_VISIBLE &&
+                             (next_window = macdrv_get_cocoa_window(next, TRUE))))
+                next = NtUserGetWindowRelative(next, GW_HWNDNEXT);
+        }
+
+        TRACE("win %p/%p below %p/%p above %p/%p\n",
+              data->hwnd, data->cocoa_window, prev, prev_window, next, next_window);
+
+        if (!prev_window)
+            activate = activate_on_focus_time && (NtGetTickCount() - activate_on_focus_time < 2000);
+        macdrv_order_cocoa_window(data->cocoa_window, prev_window, next_window, activate);
+        data->on_screen = TRUE;
+
+        info.cbSize = sizeof(info);
+        if (NtUserGetGUIThreadInfo(NtUserGetWindowThread(data->hwnd, NULL), &info) && info.hwndFocus &&
+            (data->hwnd == info.hwndFocus || NtUserIsChild(data->hwnd, info.hwndFocus)))
+            set_focus(info.hwndFocus, FALSE);
+        if (activate)
+            activate_on_focus_time = 0;
+    }
+    else
+    {
+        TRACE("win %p/%p showing view\n", data->hwnd, data->cocoa_view);
+
+        macdrv_set_view_hidden(data->cocoa_view, FALSE);
+        data->on_screen = TRUE;
+    }
 }
 
 
@@ -674,7 +811,53 @@
 
     if (data->cocoa_window)
         macdrv_hide_cocoa_window(data->cocoa_window);
+    else
+        macdrv_set_view_hidden(data->cocoa_view, TRUE);
     data->on_screen = FALSE;
+}
+
+
+/***********************************************************************
+ *              sync_window_z_order
+ */
+static void sync_window_z_order(struct macdrv_win_data *data)
+{
+    if (data->cocoa_view)
+    {
+        HWND parent = NtUserGetAncestor(data->hwnd, GA_PARENT);
+        macdrv_view superview = macdrv_get_client_cocoa_view(parent);
+        macdrv_window window = NULL;
+        HWND prev;
+        HWND next = NULL;
+        macdrv_view prev_view = NULL;
+        macdrv_view next_view = NULL;
+
+        if (!superview)
+        {
+            window = macdrv_get_cocoa_window(parent, FALSE);
+            if (!window)
+                WARN("hwnd %p/%p parent %p has no Cocoa window or view in this process\n", data->hwnd, data->cocoa_view, parent);
+        }
+
+        /* find window that this one must be after */
+        prev = NtUserGetWindowRelative(data->hwnd, GW_HWNDPREV);
+        while (prev && !(prev_view = macdrv_get_cocoa_view(prev)))
+            prev = NtUserGetWindowRelative(prev, GW_HWNDPREV);
+        if (!prev_view)
+        {
+            /* find window that this one must be before */
+            next = NtUserGetWindowRelative(data->hwnd, GW_HWNDNEXT);
+            while (next && !(next_view = macdrv_get_cocoa_view(next)))
+                next = NtUserGetWindowRelative(next, GW_HWNDNEXT);
+        }
+
+        TRACE("win %p/%p below %p/%p above %p/%p\n",
+              data->hwnd, data->cocoa_view, prev, prev_view, next, next_view);
+
+        macdrv_set_view_superview(data->cocoa_view, superview, window, prev_view, next_view);
+    }
+    else if (data->on_screen)
+        show_window(data);
 }
 
 
@@ -746,6 +929,21 @@
 
 
 /***********************************************************************
+ *              sync_client_view_position
+ */
+static void sync_client_view_position(struct macdrv_win_data *data)
+{
+    if (data->cocoa_view != data->client_cocoa_view)
+    {
+        RECT rect = data->rects.client;
+        OffsetRect(&rect, -data->rects.visible.left, -data->rects.visible.top);
+        macdrv_set_view_frame(data->client_cocoa_view, cgrect_from_rect(rect));
+        TRACE("win %p/%p client %s\n", data->hwnd, data->client_cocoa_view, wine_dbgstr_rect(&rect));
+    }
+}
+
+
+/***********************************************************************
  *              sync_window_position
  *
  * Synchronize the Mac window position with the Windows one
@@ -753,6 +951,7 @@
 static void sync_window_position(struct macdrv_win_data *data, UINT swp_flags, const struct window_rects *old_rects)
 {
     CGRect frame = cgrect_from_rect(data->rects.visible);
+    BOOL force_z_order = FALSE;
 
     if (data->cocoa_window)
     {
@@ -764,6 +963,32 @@
 
         macdrv_set_cocoa_window_frame(data->cocoa_window, &frame);
     }
+    else
+    {
+        BOOL were_equal = (data->cocoa_view == data->client_cocoa_view);
+        BOOL now_equal = EqualRect(&data->rects.visible, &data->rects.client);
+
+        if (were_equal && !now_equal)
+        {
+            data->cocoa_view = macdrv_create_view(frame);
+            macdrv_set_view_hidden(data->cocoa_view, !data->on_screen);
+            macdrv_set_view_superview(data->client_cocoa_view, data->cocoa_view, NULL, NULL, NULL);
+            macdrv_set_view_hidden(data->client_cocoa_view, FALSE);
+            force_z_order = TRUE;
+        }
+        else if (!were_equal && now_equal)
+        {
+            macdrv_dispose_view(data->cocoa_view);
+            data->cocoa_view = data->client_cocoa_view;
+            macdrv_set_view_hidden(data->cocoa_view, !data->on_screen);
+            macdrv_set_view_frame(data->cocoa_view, frame);
+            force_z_order = TRUE;
+        }
+        else if (!EqualRect(&data->rects.visible, &old_rects->visible))
+            macdrv_set_view_frame(data->cocoa_view, frame);
+    }
+
+    sync_client_view_position(data);
 
     if (old_rects &&
         (IsRectEmpty(&old_rects->window) != IsRectEmpty(&data->rects.window) ||
@@ -771,11 +996,12 @@
          old_rects->window.top - old_rects->visible.top != data->rects.window.top - data->rects.visible.top))
         sync_window_region(data, (HRGN)1);
 
-    TRACE("win %p/%p whole_rect %s frame %s\n", data->hwnd, data->cocoa_window,
+    TRACE("win %p/%p whole_rect %s frame %s\n", data->hwnd,
+          data->cocoa_window ? (void*)data->cocoa_window : (void*)data->cocoa_view,
           wine_dbgstr_rect(&data->rects.visible), wine_dbgstr_cgrect(frame));
 
-    if (data->cocoa_window && data->on_screen && (!(swp_flags & SWP_NOZORDER) || (swp_flags & SWP_SHOWWINDOW)))
-        show_window(data);
+    if (force_z_order || !(swp_flags & SWP_NOZORDER) || (swp_flags & SWP_SHOWWINDOW))
+        sync_window_z_order(data);
 }
 
 
@@ -871,7 +1097,7 @@
     capturePoint.y = (short)HIWORD(dwPoint);
     NtUserClipCursor(NULL);
 
-    TRACE("hwnd %p hittest %d, pos %d,%d\n", hwnd, hittest, capturePoint.x, capturePoint.y);
+    TRACE("hwnd %p hittest %d, pos %d,%d\n", hwnd, hittest, (int)capturePoint.x, (int)capturePoint.y);
 
     origRect.left = origRect.right = origRect.top = origRect.bottom = 0;
     if (NtUserAdjustWindowRect(&origRect, style, FALSE, NtUserGetWindowLongW(hwnd, GWL_EXSTYLE), dpi))
@@ -1070,102 +1296,6 @@
     NtUserPostMessage(hwnd, WM_SYSCOMMAND, command, 0);
 }
 
-static void macdrv_client_surface_destroy(struct client_surface *client)
-{
-    struct macdrv_client_surface *surface = impl_from_client_surface(client);
-
-    TRACE("%s\n", debugstr_client_surface(client));
-
-    if (surface->metal_view) macdrv_view_release_metal_view(surface->metal_view);
-    if (surface->metal_device) macdrv_release_metal_device(surface->metal_device);
-}
-
-static void macdrv_client_surface_detach(struct client_surface *client)
-{
-    struct macdrv_client_surface *surface = impl_from_client_surface(client);
-
-    TRACE("%s\n", debugstr_client_surface(client));
-
-    if (surface->cocoa_view)
-    {
-        struct macdrv_win_data *data;
-
-        if ((data = get_win_data(client->hwnd)))
-        {
-            if (data->client_view == surface->cocoa_view)
-                data->client_view = NULL;
-            release_win_data(data);
-        }
-
-        macdrv_dispose_view(surface->cocoa_view);
-    }
-}
-
-static void macdrv_client_surface_update(struct client_surface *client)
-{
-    struct macdrv_client_surface *surface = impl_from_client_surface(client);
-    HWND hwnd = client->hwnd, toplevel = NtUserGetAncestor(hwnd, GA_ROOT);
-    struct macdrv_win_data *data;
-    RECT rect;
-
-    TRACE("%s\n", debugstr_client_surface(client));
-
-    NtUserGetClientRect(hwnd, &rect, NtUserGetWinMonitorDpi(hwnd, MDT_RAW_DPI));
-    NtUserMapWindowPoints(hwnd, toplevel, (POINT *)&rect, 2, NtUserGetWinMonitorDpi(toplevel, MDT_RAW_DPI));
-
-    if (!(data = get_win_data(toplevel))) return;
-    OffsetRect(&rect, data->rects.client.left - data->rects.visible.left, data->rects.client.top - data->rects.visible.top);
-    macdrv_set_view_frame(surface->cocoa_view, cgrect_from_rect(rect));
-    macdrv_set_view_superview(surface->cocoa_view, toplevel == hwnd ? NULL : data->client_view, data->cocoa_window, NULL, NULL);
-    release_win_data(data);
-}
-
-static void macdrv_client_surface_present(struct client_surface *client, HDC hdc)
-{
-    struct macdrv_client_surface *surface = impl_from_client_surface(client);
-    struct macdrv_win_data *data;
-
-    TRACE("%s\n", debugstr_client_surface(client));
-
-    if (!(data = get_win_data(surface->client.hwnd))) return;
-    if (data->client_view != surface->cocoa_view)
-    {
-        if (data->client_view) macdrv_set_view_hidden(data->client_view, TRUE);
-        macdrv_set_view_hidden(surface->cocoa_view, FALSE);
-        data->client_view = surface->cocoa_view;
-    }
-    release_win_data(data);
-}
-
-static const struct client_surface_funcs macdrv_client_surface_funcs =
-{
-    .destroy = macdrv_client_surface_destroy,
-    .detach = macdrv_client_surface_detach,
-    .update = macdrv_client_surface_update,
-    .present = macdrv_client_surface_present,
-};
-
-struct macdrv_client_surface *macdrv_client_surface_create(HWND hwnd)
-{
-    HWND toplevel = NtUserGetAncestor(hwnd, GA_ROOT);
-    struct macdrv_client_surface *surface;
-    RECT rect;
-
-    NtUserGetClientRect(hwnd, &rect, NtUserGetWinMonitorDpi(hwnd, MDT_RAW_DPI));
-    NtUserMapWindowPoints(hwnd, toplevel, (POINT *)&rect, 2, NtUserGetWinMonitorDpi(toplevel, MDT_RAW_DPI));
-
-    surface = client_surface_create(sizeof(*surface), &macdrv_client_surface_funcs, hwnd);
-    surface->cocoa_view = macdrv_create_view(cgrect_from_rect(rect));
-    macdrv_set_view_hidden(surface->cocoa_view, TRUE);
-
-    if (surface)
-    {
-        macdrv_client_surface_update(&surface->client);
-        macdrv_client_surface_present(&surface->client, 0);
-    }
-
-    return surface;
-}
 
 /**********************************************************************
  *              SetDesktopWindow   (MACDRV.@)
@@ -1248,6 +1378,8 @@
     if (data->drag_event) NtSetEvent(data->drag_event, NULL);
 
     destroy_cocoa_window(data);
+    destroy_cocoa_view(data);
+    if (data->client_cocoa_view) macdrv_dispose_view(data->client_cocoa_view);
 
     CFDictionaryRemoveValue(win_datas, hwnd);
     release_win_data(data);
@@ -1281,7 +1413,7 @@
 {
     struct macdrv_win_data *data = get_win_data(hwnd);
 
-    TRACE("hwnd %p key %#08x alpha %#02x flags %x\n", hwnd, key, alpha, flags);
+    TRACE("hwnd %p key %#08x alpha %#02x flags %x\n", hwnd, (unsigned int)key, alpha, (unsigned int)flags);
 
     if (data)
     {
@@ -1319,10 +1451,14 @@
         {
             /* destroy the old Mac window */
             destroy_cocoa_window(data);
-        }
+            create_cocoa_view(data);
+        }
+
+        set_cocoa_view_parent(data, parent);
     }
     else  /* new top level window */
     {
+        destroy_cocoa_view(data);
         create_cocoa_window(data);
     }
     release_win_data(data);
@@ -1365,7 +1501,7 @@
 {
     struct macdrv_win_data *data;
 
-    TRACE("hwnd %p offset %d styleOld 0x%08x styleNew 0x%08x\n", hwnd, offset, style->styleOld, style->styleNew);
+    TRACE("hwnd %p offset %d styleOld 0x%08x styleNew 0x%08x\n", hwnd, offset, (unsigned int)style->styleOld, (unsigned int)style->styleNew);
 
     if (hwnd == NtUserGetDesktopWindow()) return;
     if (!(data = get_win_data(hwnd))) return;
@@ -1491,7 +1627,7 @@
 /***********************************************************************
  *              UpdateLayeredWindow   (MACDRV.@)
  */
-void macdrv_UpdateLayeredWindow(HWND hwnd, BYTE alpha, UINT flags)
+void macdrv_UpdateLayeredWindow(HWND hwnd, UINT flags)
 {
     struct macdrv_win_data *data;
 
@@ -1502,7 +1638,7 @@
             show_window(data);
 
         /* The ULW flags are a superset of the LWA flags. */
-        sync_window_opacity(data, alpha, TRUE, flags);
+        sync_window_opacity(data, 255, TRUE, flags);
         release_win_data(data);
     }
 }
@@ -1583,10 +1719,9 @@
 /***********************************************************************
  *              WindowPosChanged   (MACDRV.@)
  */
-void macdrv_WindowPosChanged(HWND hwnd, HWND insert_after, HWND owner_hint, UINT swp_flags,
+void macdrv_WindowPosChanged(HWND hwnd, HWND insert_after, HWND owner_hint, UINT swp_flags, BOOL fullscreen,
                              const struct window_rects *new_rects, struct window_surface *surface)
 {
-    BOOL fullscreen = swp_flags & WINE_SWP_FULLSCREEN;
     struct macdrv_thread_data *thread_data;
     struct macdrv_win_data *data;
     unsigned int new_style = NtUserGetWindowLongW(hwnd, GWL_STYLE);
@@ -1602,7 +1737,9 @@
     TRACE("win %p/%p new_rects %s style %08x flags %08x surface %p\n", hwnd, data->cocoa_window,
           debugstr_window_rects(new_rects), new_style, swp_flags, surface);
 
-    if (!data->cocoa_window) goto done;
+    sync_gl_view(data, &old_rects);
+
+    if (!data->cocoa_window && !data->cocoa_view) goto done;
 
     if (data->on_screen)
     {
@@ -1611,10 +1748,15 @@
     }
 
     /* check if we are currently processing an event relevant to this window */
-    if (!thread_data || !thread_data->current_event ||
-        !data->cocoa_window || thread_data->current_event->window != data->cocoa_window ||
-        (thread_data->current_event->type != WINDOW_FRAME_CHANGED &&
-         thread_data->current_event->type != WINDOW_DID_UNMINIMIZE))
+    if (thread_data && thread_data->current_event &&
+        data->cocoa_window && thread_data->current_event->window == data->cocoa_window &&
+        (thread_data->current_event->type == WINDOW_FRAME_CHANGED ||
+         thread_data->current_event->type == WINDOW_DID_UNMINIMIZE))
+    {
+        if (thread_data->current_event->type == WINDOW_FRAME_CHANGED)
+            sync_client_view_position(data);
+    }
+    else
     {
         sync_window_position(data, swp_flags, &old_rects);
         if (data->cocoa_window)
@@ -1633,13 +1775,8 @@
                 (data->layered || !(NtUserGetWindowLongW( hwnd, GWL_EXSTYLE ) & WS_EX_LAYERED)))
                 show_window(data);
         }
-    }
-
-    if (fullscreen || fullscreen != data->fullscreen)
-    {
-        CGRect rect = (fullscreen && !EqualRect(&data->rects.window, &data->rects.visible)) ? cgrect_from_rect(data->rects.window) : CGRectZero;
-        macdrv_set_window_mask(data->cocoa_window, rect);
-        data->fullscreen = fullscreen;
+        else if (!data->on_screen)
+            show_window(data);
     }
 
 done:
@@ -1713,16 +1850,16 @@
     if (data->rects.window.left == rect.left && data->rects.window.top == rect.top)
         flags |= SWP_NOMOVE;
     else
-        TRACE("%p moving from (%d,%d) to (%d,%d)\n", hwnd, data->rects.window.left,
-              data->rects.window.top, rect.left, rect.top);
+        TRACE("%p moving from (%d,%d) to (%d,%d)\n", hwnd, (int)data->rects.window.left,
+              (int)data->rects.window.top, (int)rect.left, (int)rect.top);
 
     if ((data->rects.window.right - data->rects.window.left == width &&
          data->rects.window.bottom - data->rects.window.top == height) ||
         (IsRectEmpty(&data->rects.window) && width == 1 && height == 1))
         flags |= SWP_NOSIZE;
     else
-        TRACE("%p resizing from (%dx%d) to (%dx%d)\n", hwnd, data->rects.window.right - data->rects.window.left,
-              data->rects.window.bottom - data->rects.window.top, width, height);
+        TRACE("%p resizing from (%dx%d) to (%dx%d)\n", hwnd, (int)(data->rects.window.right - data->rects.window.left),
+              (int)(data->rects.window.bottom - data->rects.window.top), width, height);
 
     being_dragged = data->drag_event != NULL;
     release_win_data(data);
@@ -1731,7 +1868,7 @@
         flags |= SWP_NOSENDCHANGING;
     if (!(flags & SWP_NOSIZE) || !(flags & SWP_NOMOVE))
     {
-        bool send_sizemove = !event->window_frame_changed.in_resize && !being_dragged && !event->window_frame_changed.skip_size_move_loop;
+        int send_sizemove = !event->window_frame_changed.in_resize && !being_dragged && !event->window_frame_changed.skip_size_move_loop;
         if (send_sizemove)
             send_message(hwnd, WM_ENTERSIZEMOVE, 0, 0);
         NtUserSetRawWindowPos(hwnd, rect, flags, FALSE);
@@ -1758,9 +1895,17 @@
 
     if (can_window_become_foreground(hwnd) && !(style & WS_MINIMIZE))
     {
-        TRACE("setting foreground window to %p\n", hwnd);
-        NtUserSetForegroundWindowInternal(hwnd);
-        return;
+        /* simulate a mouse click on the menu to find out
+         * whether the window wants to be activated */
+        LRESULT ma = send_message(hwnd, WM_MOUSEACTIVATE,
+                                  (WPARAM)NtUserGetAncestor(hwnd, GA_ROOT),
+                                  MAKELONG(HTMENU, WM_LBUTTONDOWN));
+        if (ma != MA_NOACTIVATEANDEAT && ma != MA_NOACTIVATE)
+        {
+            TRACE("setting foreground window to %p\n", hwnd);
+            NtUserSetForegroundWindow(hwnd);
+            return;
+        }
     }
 
     TRACE("win %p/%p rejecting focus\n", hwnd, event->window);
@@ -1783,7 +1928,7 @@
     {
         send_message(hwnd, WM_CANCELMODE, 0, 0);
         if (hwnd == NtUserGetForegroundWindow())
-            NtUserSetForegroundWindowInternal(NtUserGetDesktopWindow());
+            NtUserSetForegroundWindow(NtUserGetDesktopWindow());
     }
 }
 
@@ -1812,7 +1957,7 @@
     if (get_active_window() == NtUserGetForegroundWindow())
     {
         TRACE("setting fg to desktop\n");
-        NtUserSetForegroundWindowInternal(NtUserGetDesktopWindow());
+        NtUserSetForegroundWindow(NtUserGetDesktopWindow());
     }
 }
 
@@ -1876,6 +2021,7 @@
     {
         TRACE("restoring win %p/%p\n", hwnd, data->cocoa_window);
         release_win_data(data);
+        NtUserSetActiveWindow(hwnd);
         send_message(hwnd, WM_SYSCOMMAND, SC_RESTORE, 0);
         return;
     }
@@ -1975,7 +2121,7 @@
         if (ma != MA_NOACTIVATEANDEAT && ma != MA_NOACTIVATE)
         {
             TRACE("setting foreground window to %p\n", hwnd);
-            NtUserSetForegroundWindowInternal(hwnd);
+            NtUserSetForegroundWindow(hwnd);
         }
     }
 
