--- a/dlls/win32u/input.c
+++ b/dlls/win32u/input.c
@@ -731,27 +731,33 @@
     return ret;
 }
 
-/***********************************************************************
- *	     NtUserGetCursorPos (win32u.@)
- */
-BOOL WINAPI NtUserGetCursorPos( POINT *pt )
+static BOOL get_shared_cursor_pos( POINT *pt, DWORD *last_change )
 {
     struct object_lock lock = OBJECT_LOCK_INIT;
     const desktop_shm_t *desktop_shm;
-    BOOL ret = TRUE;
-    DWORD last_change = 0;
     NTSTATUS status;
-    RECT rect;
-
-    if (!pt) return FALSE;
 
     while ((status = get_shared_desktop( &lock, &desktop_shm )) == STATUS_PENDING)
     {
         pt->x = desktop_shm->cursor.x;
         pt->y = desktop_shm->cursor.y;
-        last_change = desktop_shm->cursor.last_change;
+        *last_change = desktop_shm->cursor.last_change;
     }
-    if (status) return FALSE;
+    return !status;
+}
+
+/***********************************************************************
+ *	     get_cursor_pos
+ */
+BOOL get_cursor_pos( POINT *pt )
+{
+    BOOL ret = TRUE;
+    DWORD last_change = 0;
+    RECT rect;
+
+    if (!pt) return FALSE;
+
+    if (!get_shared_cursor_pos( pt, &last_change )) return FALSE;
 
     /* query new position from graphics driver if we haven't updated recently */
     if (NtGetTickCount() - last_change > 100) ret = user_driver->pGetCursorPos( pt );
@@ -785,10 +791,25 @@
         info->flags = CURSOR_SHOWING;
     }
 
-    NtUserGetCursorPos( &info->ptScreenPos );
+    get_cursor_pos( &info->ptScreenPos );
     return TRUE;
 }
 
+static void check_for_events( UINT flags )
+{
+    struct peek_message_filter filter =
+    {
+        .internal = TRUE,
+        .flags = PM_REMOVE,
+    };
+    MSG msg;
+
+    if (!user_driver->pProcessEvents( flags ))
+        flush_window_surfaces( TRUE );
+
+    peek_message( &msg, &filter );
+}
+
 /**********************************************************************
  *           GetAsyncKeyState (win32u.@)
  */
@@ -799,16 +820,22 @@
     NTSTATUS status;
     BYTE state = 0;
     SHORT ret = 0;
+    static int use_recent_bit = -1;
+
+    if (use_recent_bit == -1) {
+        const char *sgi = getenv("SteamGameId");
+        use_recent_bit = sgi && (!strcmp(sgi, "302190") || !strcmp(sgi, "271760"));
+    }
 
     if (key < 0 || key >= 256) return 0;
 
-    check_for_events( QS_ALLINPUT );
+    check_for_events( QS_INPUT );
 
     while ((status = get_shared_desktop( &lock, &desktop_shm )) == STATUS_PENDING)
         state = desktop_shm->keystate[key];
 
     if (status) return 0;
-    if (!(state & 0x40)) return (state & 0x80) << 8;
+    if (!(state & 0x40) || !use_recent_bit) return (state & 0x80) << 8;
 
     /* Need to make a server call to reset the last pressed bit */
     SERVER_START_REQ( get_key_state )
@@ -989,7 +1016,7 @@
     HKL layout = thread->kbd_layout;
 
     if (thread_id && thread_id != GetCurrentThreadId())
-        FIXME( "couldn't return keyboard layout for thread %04x\n", thread_id );
+        FIXME( "couldn't return keyboard layout for thread %04x\n", (int)thread_id );
 
     if (!layout) return get_locale_kbd_layout();
     return layout;
@@ -1231,7 +1258,7 @@
     const KBDTABLES *kbd_tables;
     VSC_LPWSTR *key_name;
 
-    TRACE_(keyboard)( "lparam %#x, buffer %p, size %d.\n", lparam, buffer, size );
+    TRACE_(keyboard)( "lparam %#x, buffer %p, size %d.\n", (int)lparam, buffer, size );
 
     if (!buffer || !size) return 0;
     if ((len = user_driver->pGetKeyNameText( lparam, buffer, size )) >= 0) return len;
@@ -1267,7 +1294,7 @@
         HKL hkl = NtUserGetKeyboardLayout( 0 );
         vkey = NtUserMapVirtualKeyEx( code & 0xff, MAPVK_VSC_TO_VK, hkl );
         buffer[0] = NtUserMapVirtualKeyEx( vkey, MAPVK_VK_TO_CHAR, hkl );
-        len = buffer[0] ? 1 : 0;
+        len = 1;
     }
     buffer[len] = 0;
 
@@ -1567,7 +1594,7 @@
     unsigned int i;
 
 
-    TRACE( "%d, %p, %p, %d, %d\n", size, ptin, ptout, count, resolution );
+    TRACE( "%d, %p, %p, %d, %d\n", size, ptin, ptout, count, (int)resolution );
 
     if ((size != sizeof(MOUSEMOVEPOINT)) || (count < 0) || (count > ARRAY_SIZE( positions )))
     {
@@ -1696,7 +1723,7 @@
     POINT pos;
     HWND ret;
 
-    NtUserGetCursorPos( &pos );
+    get_cursor_pos( &pos );
     ret = window_from_point( hwnd, pos, hittest, FALSE );
     if (ret && ret == tracking->last_mouse_message_hwnd)
     {
@@ -1768,7 +1795,7 @@
     /* stop the timer if the tracking list is empty */
     if (!(tracking->info.dwFlags & (TME_HOVER | TME_LEAVE)))
     {
-        NtUserKillSystemTimer( tracking->info.hwndTrack, SYSTEM_TIMER_TRACK_MOUSE );
+        kill_system_timer( tracking->info.hwndTrack, SYSTEM_TIMER_TRACK_MOUSE );
         tracking->info.hwndTrack = 0;
         tracking->info.dwFlags = 0;
         tracking->info.dwHoverTime = 0;
@@ -1787,11 +1814,11 @@
     POINT pos;
 
     TRACE( "size %u, flags %#x, hwnd %p, time %u\n",
-           info->cbSize, info->dwFlags, info->hwndTrack, info->dwHoverTime );
+           (int)info->cbSize, (int)info->dwFlags, info->hwndTrack, (int)info->dwHoverTime );
 
     if (info->cbSize != sizeof(TRACKMOUSEEVENT))
     {
-        WARN( "wrong size %u\n", info->cbSize );
+        WARN( "wrong size %u\n", (int)info->cbSize );
         RtlSetLastWin32Error( ERROR_INVALID_PARAMETER );
         return FALSE;
     }
@@ -1821,7 +1848,7 @@
     TRACE( "point %s hwnd %p hittest %d\n", wine_dbgstr_point(&pos), hwnd, hittest );
 
     if (info->dwFlags & ~(TME_CANCEL | TME_HOVER | TME_LEAVE | TME_NONCLIENT))
-        FIXME( "ignoring flags %#x\n", info->dwFlags & ~(TME_CANCEL | TME_HOVER | TME_LEAVE | TME_NONCLIENT) );
+        FIXME( "ignoring flags %#x\n", (int)info->dwFlags & ~(TME_CANCEL | TME_HOVER | TME_LEAVE | TME_NONCLIENT) );
 
     if (info->dwFlags & TME_CANCEL)
     {
@@ -1832,7 +1859,7 @@
             /* if we aren't tracking on hover or leave remove this entry */
             if (!(tracking->info.dwFlags & (TME_HOVER | TME_LEAVE)))
             {
-                NtUserKillSystemTimer( tracking->info.hwndTrack, SYSTEM_TIMER_TRACK_MOUSE );
+                kill_system_timer( tracking->info.hwndTrack, SYSTEM_TIMER_TRACK_MOUSE );
                 tracking->info.hwndTrack = 0;
                 tracking->info.dwFlags = 0;
                 tracking->info.dwHoverTime = 0;
@@ -1841,24 +1868,13 @@
     }
     else
     {
-        /* If TME_LEAVE is set and the mouse is not in the tracking window, post WM_MOUSELEAVE
-         * now and don't overwrite the current tracking information */
-        if (info->dwFlags & TME_LEAVE && !hwnd)
-        {
-            if (info->dwFlags & TME_NONCLIENT)
-                NtUserPostMessage( info->hwndTrack, WM_NCMOUSELEAVE, 0, 0 );
-            else
-                NtUserPostMessage( info->hwndTrack, WM_MOUSELEAVE, 0, 0 );
-            return TRUE;
-        }
-
         /* In our implementation, it's possible that another window will receive
          * WM_MOUSEMOVE and call TrackMouseEvent before TrackMouseEventProc is
          * called. In such a situation, post the WM_MOUSELEAVE now. */
         if ((tracking->info.dwFlags & TME_LEAVE) && tracking->info.hwndTrack != NULL)
             check_mouse_leave( hwnd, hittest, tracking );
 
-        NtUserKillSystemTimer( tracking->info.hwndTrack, SYSTEM_TIMER_TRACK_MOUSE );
+        kill_system_timer( tracking->info.hwndTrack, SYSTEM_TIMER_TRACK_MOUSE );
         tracking->info.hwndTrack = 0;
         tracking->info.dwFlags = 0;
         tracking->info.dwHoverTime = 0;
@@ -1947,7 +1963,7 @@
  *
  * Change the focus window, sending the WM_SETFOCUS and WM_KILLFOCUS messages
  */
-static HWND set_focus_window( HWND hwnd )
+static HWND set_focus_window( HWND hwnd, BOOL from_active, BOOL force )
 {
     HWND previous = 0, ime_hwnd;
     BOOL ret;
@@ -1961,9 +1977,12 @@
     SERVER_END_REQ;
     if (!ret) return 0;
     if (previous == hwnd) return previous;
+    if (!force && hwnd == previous) return previous;
 
     if (previous)
     {
+        if (!NtUserIsWindow(hwnd) && !from_active)
+            NtUserNotifyWinEvent( EVENT_OBJECT_FOCUS, previous, OBJID_CLIENT, CHILDID_SELF );
         send_message( previous, WM_KILLFOCUS, (WPARAM)hwnd, 0 );
 
         ime_hwnd = get_default_ime_window( previous );
@@ -1980,7 +1999,8 @@
             send_message( ime_hwnd, WM_IME_INTERNAL, IME_INTERNAL_ACTIVATE,
                           HandleToUlong(hwnd) );
 
-        NtUserNotifyWinEvent( EVENT_OBJECT_FOCUS, hwnd, OBJID_CLIENT, 0 );
+        if (!from_active)
+            NtUserNotifyWinEvent( EVENT_OBJECT_FOCUS, hwnd, OBJID_CLIENT, 0 );
 
         send_message( hwnd, WM_SETFOCUS, (WPARAM)previous, 0 );
     }
@@ -1993,19 +2013,19 @@
 BOOL set_active_window( HWND hwnd, HWND *prev, BOOL mouse, BOOL focus, DWORD new_active_thread_id )
 {
     HWND previous = get_active_window();
-    BOOL ret;
+    BOOL ret = FALSE;
     DWORD old_thread, new_thread;
     CBTACTIVATESTRUCT cbt;
 
-    TRACE( "hwnd %p, previous %p, mouse %u, focus %u, new_active_thread_id %04x\n",
-           hwnd, previous, mouse, focus, new_active_thread_id );
-
     if (previous == hwnd)
     {
         if (prev) *prev = hwnd;
         goto done;
     }
 
+    if (prev) *prev = previous;
+    if (win_set_flags( hwnd, WIN_IS_ACTIVATING, 0 ) & WIN_IS_ACTIVATING) return TRUE;
+
     /* call CBT hook chain */
     cbt.fMouse     = mouse;
     cbt.hWndActive = previous;
@@ -2015,17 +2035,18 @@
     {
         send_message( previous, WM_NCACTIVATE, FALSE, (LPARAM)hwnd );
         send_message( previous, WM_ACTIVATE,
-                      MAKEWPARAM( WA_INACTIVE, is_iconic(previous) ? 0x20 : 0 ), (LPARAM)hwnd );
+                      MAKEWPARAM( WA_INACTIVE, is_iconic(previous) ), (LPARAM)hwnd );
     }
 
     SERVER_START_REQ( set_active_window )
     {
         req->handle = wine_server_user_handle( hwnd );
+        req->internal_msg = WM_WINE_SETACTIVEWINDOW;
         if ((ret = !wine_server_call_err( req )))
             previous = wine_server_ptr_handle( reply->previous );
     }
     SERVER_END_REQ;
-    if (!ret) return FALSE;
+    if (!ret) goto done;
     if (prev) *prev = previous;
     if (previous == hwnd) goto done;
 
@@ -2037,7 +2058,7 @@
         if (send_message( hwnd, WM_QUERYNEWPALETTE, 0, 0 ))
             send_message_timeout( HWND_BROADCAST, WM_PALETTEISCHANGING, (WPARAM)hwnd, 0,
                                   SMTO_ABORTIFHUNG, 2000, FALSE );
-        if (!is_window(hwnd)) return FALSE;
+        if (!(ret = is_window(hwnd))) goto done;
     }
 
     old_thread = previous ? get_window_thread( previous, NULL ) : 0;
@@ -2072,9 +2093,11 @@
 
     if (is_window(hwnd))
     {
-        send_message( hwnd, WM_NCACTIVATE, hwnd == NtUserGetForegroundWindow(), (LPARAM)previous );
+        send_message( hwnd, WM_NCACTIVATE,
+                      (hwnd == NtUserGetForegroundWindow()) && !(win_get_flags(previous) & WIN_IS_ACTIVATING),
+                      (LPARAM)previous );
         send_message( hwnd, WM_ACTIVATE,
-                      MAKEWPARAM( mouse ? WA_CLICKACTIVE : WA_ACTIVE, is_iconic(hwnd) ? 0x20 : 0 ),
+                      MAKEWPARAM( mouse ? WA_CLICKACTIVE : WA_ACTIVE, is_iconic(hwnd) ),
                       (LPARAM)previous );
         if (NtUserGetAncestor( hwnd, GA_PARENT ) == get_desktop_window())
             NtUserPostMessage( get_desktop_window(), WM_PARENTNOTIFY, WM_NCACTIVATE, (LPARAM)hwnd );
@@ -2091,12 +2114,13 @@
         if (hwnd == info.hwndActive)
         {
             if (!info.hwndFocus || !hwnd || NtUserGetAncestor( info.hwndFocus, GA_ROOT ) != hwnd)
-                set_focus_window( hwnd );
+                set_focus_window( hwnd, TRUE, FALSE );
         }
     }
 
 done:
-    if (hwnd)
+    win_set_flags( hwnd, 0, WIN_IS_ACTIVATING );
+    if (ret && hwnd)
     {
         if (hwnd == NtUserGetForegroundWindow()) user_driver->pActivateWindow( hwnd, previous );
         clip_fullscreen_window( hwnd, FALSE );
@@ -2115,7 +2139,7 @@
 
     if (hwnd)
     {
-        DWORD style;
+        LONG style;
 
         hwnd = get_full_window_handle( hwnd );
         if (!is_window( hwnd ))
@@ -2156,7 +2180,7 @@
         for (;;)
         {
             HWND parent;
-            DWORD style = get_window_long( hwndTop, GWL_STYLE );
+            LONG style = get_window_long( hwndTop, GWL_STYLE );
             if (style & (WS_MINIMIZE | WS_DISABLED)) return 0;
             if (!(style & WS_CHILD)) break;
             parent = NtUserGetAncestor( hwndTop, GA_PARENT );
@@ -2173,7 +2197,7 @@
         if (call_hooks( WH_CBT, HCBT_SETFOCUS, (WPARAM)hwnd, (LPARAM)previous, 0 )) return 0;
 
         /* activate hwndTop if needed. */
-        if (!(active = get_active_window()) && !set_foreground_window( hwndTop, FALSE, FALSE )) return 0;
+        if (!(active = get_active_window()) && !set_foreground_window( hwndTop, FALSE )) return 0;
         if (hwndTop != active)
         {
             if (!set_active_window( hwndTop, NULL, FALSE, FALSE, 0 )) return 0;
@@ -2190,21 +2214,13 @@
     }
 
     /* change focus and send messages */
-    return set_focus_window( hwnd );
-}
-
-/*****************************************************************
- *           NtUserSetForegroundWindow  (win32u.@)
- */
-BOOL WINAPI NtUserSetForegroundWindow( HWND hwnd )
-{
-    return set_foreground_window( hwnd, FALSE, FALSE );
+    return set_focus_window( hwnd, FALSE, hwnd != previous );
 }
 
 /*******************************************************************
  *		set_foreground_window
  */
-BOOL set_foreground_window( HWND hwnd, BOOL mouse, BOOL internal )
+BOOL set_foreground_window( HWND hwnd, BOOL mouse )
 {
     BOOL ret, send_msg_old = FALSE, send_msg_new = FALSE;
     DWORD new_thread_id;
@@ -2216,7 +2232,6 @@
     SERVER_START_REQ( set_foreground_window )
     {
         req->handle = wine_server_user_handle( hwnd );
-        req->internal = internal;
         if ((ret = !wine_server_call_err( req )))
         {
             previous = wine_server_ptr_handle( reply->previous );
@@ -2226,19 +2241,22 @@
     }
     SERVER_END_REQ;
 
-    TRACE( "hwnd %p, mouse %u, internal %u -> ret %u, previous %p\n", hwnd, mouse, internal, ret, previous );
-
     if (ret && previous != hwnd)
     {
         if (send_msg_old)  /* old window belongs to other thread */
-            NtUserMessageCall( previous, WM_WINE_SETACTIVEWINDOW, 0, new_thread_id,
-                               0, NtUserSendNotifyMessage, FALSE );
+            NtUserPostMessage( previous, WM_WINE_SETACTIVEWINDOW, 0, new_thread_id );
         else if (send_msg_new)  /* old window belongs to us but new one to other thread */
             ret = set_active_window( 0, NULL, mouse, TRUE, new_thread_id );
 
+        /* already active, set_active_window will do no nothing */
+        if (!send_msg_new && hwnd == get_active_window())
+        {
+            send_message( hwnd, WM_NCACTIVATE, TRUE, (LPARAM)hwnd );
+            NtUserNotifyWinEvent( EVENT_SYSTEM_FOREGROUND, hwnd, 0, 0 );
+        }
+
         if (send_msg_new)  /* new window belongs to other thread */
-            NtUserMessageCall( hwnd, WM_WINE_SETACTIVEWINDOW, (WPARAM)hwnd, 0,
-                               0, NtUserSendNotifyMessage, FALSE );
+            NtUserPostMessage( hwnd, WM_WINE_SETACTIVEWINDOW, (WPARAM)hwnd, 0 );
         else  /* new window belongs to us */
             ret = set_active_window( hwnd, NULL, mouse, TRUE, 0 );
     }
@@ -2369,7 +2387,7 @@
     if (prev && !hidden)  /* hide the previous one */
     {
         /* FIXME: won't work if prev belongs to a different process */
-        NtUserKillSystemTimer( prev, SYSTEM_TIMER_CARET );
+        kill_system_timer( prev, SYSTEM_TIMER_CARET );
         if (old_state) display_caret( prev, &r );
     }
 
@@ -2408,7 +2426,7 @@
     if (ret && prev && !hidden)
     {
         /* FIXME: won't work if prev belongs to a different process */
-        NtUserKillSystemTimer( prev, SYSTEM_TIMER_CARET );
+        kill_system_timer( prev, SYSTEM_TIMER_CARET );
         if (old_state) display_caret( prev, &r );
     }
     if (caret.bitmap) NtGdiDeleteObjectApp( caret.bitmap );
@@ -2425,9 +2443,9 @@
 }
 
 /*******************************************************************
- *           NtUserSetCaretBlinkTime  (win32u.@)
+ *              set_caret_blink_time
  */
-BOOL WINAPI NtUserSetCaretBlinkTime( unsigned int time )
+BOOL set_caret_blink_time( unsigned int time )
 {
     TRACE( "time %u\n", time );
 
@@ -2469,10 +2487,10 @@
     return user_driver->pSetIMECompositionRect( NtUserGetAncestor( hwnd, GA_ROOT ), rect );
 }
 
-/*****************************************************************
- *           NtUserSetCaretPos  (win32u.@)
+/*******************************************************************
+ *              set_caret_pos
  */
-BOOL WINAPI NtUserSetCaretPos( INT x, INT y )
+BOOL set_caret_pos( int x, int y )
 {
     int old_state = 0;
     int hidden = 0;
@@ -2579,7 +2597,7 @@
     if (ret && !hidden)
     {
         if (old_state) display_caret( hwnd, &r );
-        NtUserKillSystemTimer( hwnd, SYSTEM_TIMER_CARET );
+        kill_system_timer( hwnd, SYSTEM_TIMER_CARET );
     }
     return ret;
 }
@@ -2641,6 +2659,26 @@
     return enable_mouse_in_pointer;
 }
 
+void update_mouse_state_from_pointer( HWND hwnd, UINT msg, unsigned int pointer_id )
+{
+    BOOL update_cursor_pos = FALSE;
+    DWORD last_change;
+    POINT pt = { 0 };
+
+    SERVER_START_REQ( track_mouse_from_pointer )
+    {
+        req->win = wine_server_user_handle( hwnd );
+        req->msg = msg;
+        req->pointer_id = pointer_id;
+        if (!wine_server_call( req )) update_cursor_pos = reply->cursor_pos_updated;
+    }
+    SERVER_END_REQ;
+
+    if (!update_cursor_pos) return;
+    if (get_shared_cursor_pos( &pt, &last_change ))
+        user_driver->pSetCursorPos( pt.x, pt.y );
+}
+
 static BOOL is_captured_by_system(void)
 {
     GUITHREADINFO info;
@@ -2648,12 +2686,6 @@
     return NtUserGetGUIThreadInfo( GetCurrentThreadId(), &info ) && info.hwndCapture && (info.flags & (GUI_INMOVESIZE | GUI_INMENUMODE));
 }
 
-static BOOL is_fullscreen( const MONITORINFO *info, const RECT *rect )
-{
-    return rect->left <= info->rcMonitor.left && rect->right >= info->rcMonitor.right &&
-           rect->top <= info->rcMonitor.top && rect->bottom >= info->rcMonitor.bottom;
-}
-
 /***********************************************************************
  *      clip_fullscreen_window
  *
@@ -2663,9 +2695,9 @@
 {
     struct user_thread_info *thread_info = get_user_thread_info();
     MONITORINFO monitor_info = {.cbSize = sizeof(MONITORINFO)};
-    RECT monitor_rect, window_rect, virtual_rect;
+    RECT rect, virtual_rect;
     DWORD style;
-    UINT ctx;
+    UINT dpi, ctx;
     BOOL ret;
 
     if (hwnd == NtUserGetDesktopWindow()) return FALSE;
@@ -2677,19 +2709,19 @@
     /* maximized windows don't count as full screen */
     if ((style & WS_MAXIMIZE) && (style & WS_CAPTION) == WS_CAPTION) return FALSE;
 
+    dpi = get_dpi_for_window( hwnd );
+    if (!get_window_rect( hwnd, &rect, dpi )) return FALSE;
+    if (!is_window_rect_full_screen( &rect, dpi )) return FALSE;
+    if (is_captured_by_system()) return FALSE;
+    if (NtGetTickCount() - thread_info->clipping_reset < 1000) return FALSE;
+    if (!reset && clipping_cursor && thread_info->clipping_cursor) return FALSE;  /* already clipping */
+
     ctx = set_thread_dpi_awareness_context( NTUSER_DPI_PER_MONITOR_AWARE );
-    if (!(ret = get_present_rect( hwnd, &window_rect, get_thread_dpi() )))
-        ret = get_window_rect( hwnd, &window_rect, get_thread_dpi() );
     monitor_info = monitor_info_from_window( hwnd, MONITOR_DEFAULTTONEAREST );
     virtual_rect = get_virtual_screen_rect( get_thread_dpi(), MDT_DEFAULT );
-    monitor_rect = map_rect_virt_to_raw( monitor_info.rcMonitor, get_thread_dpi() );
+    rect = map_rect_virt_to_raw( monitor_info.rcMonitor, get_thread_dpi() );
     set_thread_dpi_awareness_context( ctx );
 
-    if (!ret || !is_fullscreen( &monitor_info, &window_rect )) return FALSE;
-    if (is_captured_by_system()) return FALSE;
-    if (NtGetTickCount() - thread_info->clipping_reset < 1000) return FALSE;
-    if (!reset && clipping_cursor && thread_info->clipping_cursor) return FALSE;  /* already clipping */
-
     if (!grab_fullscreen)
     {
         if (!EqualRect( &monitor_info.rcMonitor, &virtual_rect )) return FALSE;
@@ -2701,7 +2733,7 @@
     SERVER_START_REQ( set_cursor )
     {
         req->flags = SET_CURSOR_CLIP | SET_CURSOR_FSCLIP;
-        req->clip  = wine_server_rectangle( monitor_rect );
+        req->clip  = wine_server_rectangle( rect );
         ret = !wine_server_call( req );
     }
     SERVER_END_REQ;
@@ -2710,18 +2742,44 @@
 }
 
 /**********************************************************************
+ *       NtUserIsTouchWindow    (win32u.@)
+ */
+BOOL WINAPI NtUserIsTouchWindow( HWND hwnd, ULONG *flags )
+{
+    DWORD win_flags = win_set_flags( hwnd, 0, 0 );
+    TRACE( "hwnd %p, flags %p.\n", hwnd, flags );
+    return (win_flags & WIN_IS_TOUCH) != 0;
+}
+
+
+BOOL register_touch_window( HWND hwnd, UINT flags )
+{
+    DWORD win_flags = win_set_flags( hwnd, WIN_IS_TOUCH, 0 );
+    TRACE( "hwnd %p, flags %#x.\n", hwnd, flags );
+    return (win_flags & WIN_IS_TOUCH) == 0;
+}
+
+
+BOOL unregister_touch_window( HWND hwnd )
+{
+    DWORD win_flags = win_set_flags( hwnd, 0, WIN_IS_TOUCH );
+    TRACE( "hwnd %p.\n", hwnd );
+    return (win_flags & WIN_IS_TOUCH) != 0;
+}
+
+/**********************************************************************
  *       NtUserGetPointerInfoList    (win32u.@)
  */
 BOOL WINAPI NtUserGetPointerInfoList( UINT32 id, POINTER_INPUT_TYPE type, UINT_PTR unk0, UINT_PTR unk1, SIZE_T size,
                                       UINT32 *entry_count, UINT32 *pointer_count, void *pointer_info )
 {
-    FIXME( "id %#x, type %#x, unk0 %#lx, unk1 %#lx, size %#lx, entry_count %p, pointer_count %p, pointer_info %p stub!\n",
-           id, type, (long)unk0, (long)unk1, size, entry_count, pointer_count, pointer_info );
+    FIXME( "id %#x, type %#x, unk0 %#zx, unk1 %#zx, size %#zx, entry_count %p, pointer_count %p, pointer_info %p stub!\n",
+           id, (int)type, (size_t)unk0, (size_t)unk1, (size_t)size, entry_count, pointer_count, pointer_info );
     RtlSetLastWin32Error( ERROR_CALL_NOT_IMPLEMENTED );
     return FALSE;
 }
 
-static BOOL get_clip_cursor( RECT *rect, UINT dpi, MONITOR_DPI_TYPE type )
+BOOL get_clip_cursor( RECT *rect, UINT dpi, MONITOR_DPI_TYPE type )
 {
     struct object_lock lock = OBJECT_LOCK_INIT;
     const desktop_shm_t *desktop_shm;
@@ -2772,35 +2830,46 @@
     return TRUE;
 }
 
-/**********************************************************************
- *       NtUserGetClipCursor (win32u.@)
- */
-BOOL WINAPI NtUserGetClipCursor( RECT *rect )
-{
-    return get_clip_cursor( rect, get_thread_dpi(), MDT_DEFAULT );
-}
-
 /***********************************************************************
  *       NtUserClipCursor (win32u.@)
  */
 BOOL WINAPI NtUserClipCursor( const RECT *rect )
 {
-    UINT dpi = get_thread_dpi();
-    RECT new_rect;
+    static int keep_inside_window = -1;
+    HWND foreground = NtUserGetForegroundWindow();
+    RECT new_rect, full_rect;
     BOOL ret;
 
     TRACE( "Clipping to %s\n", wine_dbgstr_rect(rect) );
 
+    if (foreground == NtUserGetDesktopWindow())
+    {
+        WARN( "desktop is foreground, ignoring ClipCursor\n" );
+        rect = NULL;
+    }
+
     if (rect)
     {
-        HWND foreground = NtUserGetForegroundWindow();
-        if (IsRectEmpty( rect ) && get_present_rect( foreground, &new_rect, dpi ))
+        if (rect->left > rect->right || rect->top > rect->bottom) return FALSE;
+
+        if (keep_inside_window == -1)
         {
-            WARN( "Fullscreen clipping fixup to %s\n", wine_dbgstr_rect(&new_rect) );
-            rect = &new_rect;
+            const char *sgi = getenv( "SteamGameId" );
+            keep_inside_window = sgi && !strcmp( sgi, "730830" ); /* Escape from Monkey Island */
         }
-        if (rect->left > rect->right || rect->top > rect->bottom) return FALSE;
-        new_rect = map_rect_virt_to_raw( *rect, dpi );
+
+        /* keep the mouse clipped inside of a fullscreen foreground window */
+        if (keep_inside_window && NtUserGetWindowRect( foreground, &full_rect, get_thread_dpi() ) &&
+            is_window_rect_full_screen( &full_rect, get_thread_dpi() ))
+        {
+            full_rect.left = max( full_rect.left, min( full_rect.right - 1, rect->left ) );
+            full_rect.right = max( full_rect.left, min( full_rect.right - 1, rect->right ) );
+            full_rect.top = max( full_rect.top, min( full_rect.bottom - 1, rect->top ) );
+            full_rect.bottom = max( full_rect.top, min( full_rect.bottom - 1, rect->bottom ) );
+            rect = &full_rect;
+        }
+
+        new_rect = map_rect_virt_to_raw( *rect, get_thread_dpi() );
         rect = &new_rect;
     }
 
@@ -2830,15 +2899,24 @@
     return FALSE;
 }
 
-/**********************************************************************
- *       NtUserScheduleDispatchNotification    (win32u.@)
+/*****************************************************************************
+ * NtUserGetTouchInputInfo (WIN32U.@)
  */
-INT WINAPI NtUserScheduleDispatchNotification( HWND hwnd )
+BOOL WINAPI NtUserGetTouchInputInfo( HTOUCHINPUT handle, UINT count, TOUCHINPUT *ptr, int size )
 {
-    FIXME("hwnd %p stub!\n", hwnd);
+    struct user_thread_info *thread_info = get_user_thread_info();
+    struct touchinput_thread_data *thread_data;
+    UINT index = (ULONG_PTR)handle;
 
-    if (is_window(hwnd))
-        return 2;
+    TRACE( "handle %p, count %u, ptr %p, size %u.\n", handle, count, ptr, size );
 
-    return 0;
+    if (!thread_info || !(thread_data = thread_info->touchinput) || size != sizeof(TOUCHINPUT) ||
+        index >= ARRAY_SIZE(thread_data->history))
+    {
+        RtlSetLastWin32Error( ERROR_INVALID_PARAMETER );
+        return FALSE;
+    }
+
+    memcpy( ptr, thread_data->history + index, min( count, ARRAY_SIZE(thread_data->current) ) * size );
+    return TRUE;
 }
