diff --git a/dlls/d3dx9_36/tests/texture.c b/dlls/d3dx9_36/tests/texture.c
--- a/dlls/d3dx9_36/tests/texture.c
+++ b/dlls/d3dx9_36/tests/texture.c
@@ -177,23 +177,6 @@
             && compare_uint((c1 >> 24) & 0xff, (c2 >> 24) & 0xff, max_diff);
 }
 
-static uint8_t pixel_4bpp_max_channel_diff(uint32_t c1, uint32_t c2)
-{
-    uint8_t max_diff = 0;
-    unsigned int i;
-
-    for (i = 0; i < 4; ++i)
-    {
-        uint8_t x, y;
-
-        x = (c1 >> (i * 8)) & 0xff;
-        y = (c2 >> (i * 8)) & 0xff;
-        max_diff = max(max_diff, (x > y ? x - y : y - x));
-    }
-
-    return max_diff;
-}
-
 struct surface_readback
 {
     IDirect3DSurface9 *surface;
@@ -628,18 +611,6 @@
         ok(format == D3DFMT_A8R8G8B8, "Got unexpected format %u.\n", format);
     }
 
-    /* D3DFMT CxV8U8, unsupported by modern cards, replaced with D3DFMT_X8L8V8U8. */
-    if (SUCCEEDED(IDirect3D9_CheckDeviceFormat(d3d, params.AdapterOrdinal, params.DeviceType,
-                                               mode.Format, 0, D3DRTYPE_TEXTURE, D3DFMT_X8L8V8U8)))
-        expected = D3DFMT_X8L8V8U8;
-    else
-        expected = D3DFMT_V8U8;
-
-    format = D3DFMT_CxV8U8;
-    hr = D3DXCheckTextureRequirements(device, NULL, NULL, NULL, 0, &format, D3DPOOL_DEFAULT);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    ok(format == expected, "Unexpected format %u, expected %u.\n", format, expected);
-
     IDirect3D9_Release(d3d);
 }
 
@@ -1077,14 +1048,8 @@
     IDirect3DTexture9 *tex;
     IDirect3DCubeTexture9 *cubetex;
     IDirect3DVolumeTexture9 *voltex;
-    D3DLOCKED_RECT lock_rect;
-    IDirect3DSurface9 *surf;
-    D3DLOCKED_BOX lock_box;
-    IDirect3DVolume9 *vol;
-    D3DBOX box;
     HRESULT hr;
     uint32_t i;
-    RECT rect;
 
     hr = IDirect3DDevice9_CreateTexture(device, 256, 256, 5, 0, D3DFMT_A8R8G8B8, D3DPOOL_MANAGED, &tex, NULL);
 
@@ -1226,137 +1191,6 @@
     }
     else
         skip("Failed to create volume texture\n");
-
-    /* Test D3DXFilterTexture() D3DX_DEFAULT behavior. */
-    hr = IDirect3DDevice9_CreateTexture(device, 16, 16, 2, 0, D3DFMT_A8R8G8B8, D3DPOOL_SCRATCH, &tex, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DTexture9_GetSurfaceLevel(tex, 0, &surf);
-    SetRect(&rect, 0, 0, 16, 16);
-
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, a8r8g8b8_16_16, D3DFMT_A8R8G8B8, 64, NULL,
-            &rect, D3DX_FILTER_NONE, 0);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DSurface9_Release(surf);
-
-    /* Default matches D3DX_FILTER_BOX. */
-    hr = D3DXFilterTexture((IDirect3DBaseTexture9 *)tex, NULL, 0, D3DX_DEFAULT);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    IDirect3DTexture9_LockRect(tex, 1, &lock_rect, NULL, D3DLOCK_READONLY);
-    check_test_readback(lock_rect.pBits, lock_rect.Pitch, 0, a8r8g8b8_16_16_linear_filter_8_8, 8, 8, 1,
-            D3DFMT_A8R8G8B8, 0);
-    IDirect3DTexture9_UnlockRect(tex, 1);
-    IDirect3DTexture9_Release(tex);
-
-    /* Test with a non power of 2 texture size. */
-    hr = IDirect3DDevice9_CreateTexture(device, 6, 6, 2, 0, D3DFMT_A8R8G8B8, D3DPOOL_SCRATCH, &tex, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    IDirect3DTexture9_GetSurfaceLevel(tex, 0, &surf);
-    SetRect(&rect, 0, 0, 6, 6);
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, a8r8g8b8_6_6, D3DFMT_A8R8G8B8, 24, NULL,
-            &rect, D3DX_FILTER_NONE, 0);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DSurface9_Release(surf);
-
-    /*
-     * MSDN claims non power of 2 textures default to
-     * D3DX_FILTER_BOX | D3DX_FILTER_DITHER, which doesn't make much sense.
-     * It's actually D3DX_FILTER_TRIANGLE | D3DX_FILTER_DITHER, which matches
-     * the behavior of using D3DX_DEFAULT as the filter argument for all other
-     * functions.
-     */
-    hr = D3DXFilterTexture((IDirect3DBaseTexture9 *)tex, NULL, 0, D3DX_DEFAULT);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    IDirect3DTexture9_LockRect(tex, 1, &lock_rect, NULL, D3DLOCK_READONLY);
-    todo_wine check_test_readback(lock_rect.pBits, lock_rect.Pitch, 0, a8r8g8b8_6_6_triangle_dither_filter_3_3, 3, 3, 1,
-            D3DFMT_A8R8G8B8, 0);
-    IDirect3DTexture9_UnlockRect(tex, 1);
-
-    /*
-     * Do the same thing as D3DXFilterTexture() using D3DXLoadSurfaceFromMemory()
-     * to confirm behavior.
-     */
-    IDirect3DTexture9_GetSurfaceLevel(tex, 1, &surf);
-    SetRect(&rect, 0, 0, 6, 6);
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, a8r8g8b8_6_6, D3DFMT_A8R8G8B8, 24, NULL,
-            &rect, D3DX_FILTER_TRIANGLE | D3DX_FILTER_DITHER, 0);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DSurface9_Release(surf);
-
-    IDirect3DTexture9_LockRect(tex, 1, &lock_rect, NULL, D3DLOCK_READONLY);
-    todo_wine check_test_readback(lock_rect.pBits, lock_rect.Pitch, 0, a8r8g8b8_6_6_triangle_dither_filter_3_3, 3, 3, 1,
-            D3DFMT_A8R8G8B8, 0);
-    IDirect3DTexture9_UnlockRect(tex, 1);
-
-    IDirect3DTexture9_Release(tex);
-
-    /* D3DX_DEFAULT behavior for volume textures. */
-    hr = IDirect3DDevice9_CreateVolumeTexture(device, 8, 8, 8, 0, 0, D3DFMT_A8R8G8B8, D3DPOOL_SCRATCH, &voltex, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DVolumeTexture9_GetVolumeLevel(voltex, 0, &vol);
-    set_box(&box, 0, 0, 8, 8, 0, 8);
-
-    hr = D3DXLoadVolumeFromMemory(vol, NULL, NULL, a8r8g8b8_8_8_8, D3DFMT_A8R8G8B8, 32, 256, NULL, &box,
-            D3DX_FILTER_NONE, 0);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DVolume9_Release(vol);
-
-    /* Default matches D3DX_FILTER_BOX. */
-    hr = D3DXFilterTexture((IDirect3DBaseTexture9 *)voltex, NULL, 0, D3DX_DEFAULT);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    IDirect3DVolumeTexture9_LockBox(voltex, 1, &lock_box, NULL, D3DLOCK_READONLY);
-    if (sizeof(void *) == 4)
-    {
-        check_test_readback(lock_box.pBits, lock_box.RowPitch, lock_box.SlicePitch,
-                a8r8g8b8_8_8_8_box_filter_4_4_4_32bit, 4, 4, 4, D3DFMT_A8R8G8B8, 1);
-    }
-    else
-    {
-        check_test_readback(lock_box.pBits, lock_box.RowPitch, lock_box.SlicePitch,
-                a8r8g8b8_8_8_8_linear_filter_4_4_4, 4, 4, 4, D3DFMT_A8R8G8B8, 0);
-    }
-    IDirect3DVolumeTexture9_UnlockBox(voltex, 1);
-    IDirect3DVolumeTexture9_Release(voltex);
-
-    /* Test with a non power of 2 volume texture size. */
-    hr = IDirect3DDevice9_CreateVolumeTexture(device, 6, 6, 6, 0, 0, D3DFMT_A8R8G8B8, D3DPOOL_SCRATCH, &voltex, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DVolumeTexture9_GetVolumeLevel(voltex, 0, &vol);
-    set_box(&box, 0, 0, 6, 6, 0, 6);
-
-    hr = D3DXLoadVolumeFromMemory(vol, NULL, NULL, a8r8g8b8_6_6_6, D3DFMT_A8R8G8B8, 24, 144, NULL, &box,
-            D3DX_FILTER_NONE, 0);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DVolume9_Release(vol);
-
-    /*
-     * Default on non power of 2 volume texture, matches
-     * D3DX_FILTER_TRIANGLE | D3DX_FILTER_DITHER.
-     */
-    hr = D3DXFilterTexture((IDirect3DBaseTexture9 *)voltex, NULL, 0, D3DX_DEFAULT);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    IDirect3DVolumeTexture9_LockBox(voltex, 1, &lock_box, NULL, D3DLOCK_READONLY);
-    todo_wine check_test_readback(lock_box.pBits, lock_box.RowPitch, lock_box.SlicePitch,
-            a8r8g8b8_6_6_6_triangle_dither_filter_3_3_3, 3, 3, 3, D3DFMT_A8R8G8B8, 0);
-    IDirect3DVolumeTexture9_UnlockBox(voltex, 1);
-
-    IDirect3DVolumeTexture9_GetVolumeLevel(voltex, 1, &vol);
-    set_box(&box, 0, 0, 6, 6, 0, 6);
-
-    hr = D3DXLoadVolumeFromMemory(vol, NULL, NULL, a8r8g8b8_6_6_6, D3DFMT_A8R8G8B8, 24, 144, NULL, &box,
-            D3DX_FILTER_TRIANGLE | D3DX_FILTER_DITHER, 0);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    IDirect3DVolume9_Release(vol);
-
-    IDirect3DVolumeTexture9_LockBox(voltex, 1, &lock_box, NULL, D3DLOCK_READONLY);
-    todo_wine check_test_readback(lock_box.pBits, lock_box.RowPitch, lock_box.SlicePitch,
-            a8r8g8b8_6_6_6_triangle_dither_filter_3_3_3, 3, 3, 3, D3DFMT_A8R8G8B8, 0);
-    IDirect3DVolumeTexture9_UnlockBox(voltex, 1);
-    IDirect3DVolumeTexture9_Release(voltex);
 
     /* Test textures with D3DUSAGE_AUTOGENMIPMAP usage */
     if (!is_autogenmipmap_supported(device, D3DRTYPE_TEXTURE))
@@ -2321,15 +2155,13 @@
     {
         0xffff0000, 0xff00ff00, 0xff0000ff,
     };
+    HRESULT hr;
     struct surface_readback surface_rb;
     uint32_t miplevels, mip_level, i;
     IDirect3DTexture9 *texture;
     IDirect3DSurface9 *surface;
     D3DXIMAGE_INFO img_info;
     D3DSURFACE_DESC desc;
-    D3DFORMAT expected;
-    IDirect3D9 *d3d;
-    HRESULT hr;
 
     hr = D3DXCreateTextureFromFileInMemoryEx(device, dds_16bit, sizeof(dds_16bit), D3DX_DEFAULT, D3DX_DEFAULT, D3DX_DEFAULT,
         0, D3DFMT_UNKNOWN, D3DPOOL_DEFAULT, D3DX_DEFAULT, D3DX_DEFAULT, 0, NULL, NULL, &texture);
@@ -2409,10 +2241,7 @@
     ok(hr == D3D_OK, "D3DXCreateTextureFromFileInMemoryEx returned %#lx, expected %#lx.\n", hr, D3D_OK);
     IDirect3DTexture9_GetSurfaceLevel(texture, 0, &surface);
     IDirect3DSurface9_GetDesc(surface, &desc);
-    IDirect3DDevice9_GetDirect3D(device, &d3d);
-    expected = SUCCEEDED(IDirect3D9_CheckDeviceFormat(d3d, D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL,
-            D3DFMT_X8R8G8B8, 0, D3DRTYPE_TEXTURE, D3DFMT_X1R5G5B5)) ? D3DFMT_X1R5G5B5 : D3DFMT_R5G6B5;
-    ok(desc.Format == expected, "Returned format %u, expected %u.\n", desc.Format, expected);
+    ok(desc.Format == D3DFMT_X1R5G5B5, "Returned format %u, expected %u.\n", desc.Format, D3DFMT_X1R5G5B5);
     IDirect3DSurface9_Release(surface);
     IDirect3DTexture9_Release(texture);
     hr = D3DXCreateTextureFromFileInMemoryEx(device, dds_16bit, sizeof(dds_16bit),
@@ -2421,9 +2250,7 @@
     ok(hr == D3D_OK, "D3DXCreateTextureFromFileInMemoryEx returned %#lx, expected %#lx.\n", hr, D3D_OK);
     IDirect3DTexture9_GetSurfaceLevel(texture, 0, &surface);
     IDirect3DSurface9_GetDesc(surface, &desc);
-    expected = SUCCEEDED(IDirect3D9_CheckDeviceFormat(d3d, D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL,
-            D3DFMT_X8R8G8B8, 0, D3DRTYPE_TEXTURE, D3DFMT_A1R5G5B5)) ? D3DFMT_A1R5G5B5 : D3DFMT_R5G6B5;
-    ok(desc.Format == expected, "Returned format %u, expected %u.\n", desc.Format, expected);
+    ok(desc.Format == D3DFMT_A1R5G5B5, "Returned format %u, expected %u.\n", desc.Format, D3DFMT_A1R5G5B5);
     IDirect3DSurface9_Release(surface);
     IDirect3DTexture9_Release(texture);
     hr = D3DXCreateTextureFromFileInMemoryEx(device, dds_16bit, sizeof(dds_16bit),
@@ -2432,9 +2259,7 @@
     ok(hr == D3D_OK, "D3DXCreateTextureFromFileInMemoryEx returned %#lx, expected %#lx.\n", hr, D3D_OK);
     IDirect3DTexture9_GetSurfaceLevel(texture, 0, &surface);
     IDirect3DSurface9_GetDesc(surface, &desc);
-    expected = SUCCEEDED(IDirect3D9_CheckDeviceFormat(d3d, D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL,
-            D3DFMT_X8R8G8B8, 0, D3DRTYPE_TEXTURE, D3DFMT_X1R5G5B5)) ? D3DFMT_X1R5G5B5 : D3DFMT_R5G6B5;
-    ok(desc.Format == expected, "Returned format %u, expected %u.\n", desc.Format, expected);
+    ok(desc.Format == D3DFMT_X1R5G5B5, "Returned format %u, expected %u.\n", desc.Format, D3DFMT_X1R5G5B5);
     IDirect3DSurface9_Release(surface);
     IDirect3DTexture9_Release(texture);
 
@@ -2472,9 +2297,7 @@
     ok(hr == D3D_OK, "D3DXCreateTextureFromFileInMemoryEx returned %#lx, expected %#lx.\n", hr, D3D_OK);
     IDirect3DTexture9_GetSurfaceLevel(texture, 0, &surface);
     IDirect3DSurface9_GetDesc(surface, &desc);
-    expected = SUCCEEDED(IDirect3D9_CheckDeviceFormat(d3d, D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL,
-            D3DFMT_X8R8G8B8, 0, D3DRTYPE_TEXTURE, D3DFMT_L8)) ? D3DFMT_L8 : D3DFMT_X8R8G8B8;
-    ok(desc.Format == expected, "Returned format %u, expected %u.\n", desc.Format, expected);
+    ok(desc.Format == D3DFMT_L8, "Returned format %u, expected %u.\n", desc.Format, D3DFMT_L8);
     IDirect3DSurface9_Release(surface);
     IDirect3DTexture9_Release(texture);
     hr = D3DXCreateTextureFromFileInMemoryEx(device, png_grayscale, sizeof(png_grayscale),
@@ -2483,9 +2306,7 @@
     ok(hr == D3D_OK, "D3DXCreateTextureFromFileInMemoryEx returned %#lx, expected %#lx.\n", hr, D3D_OK);
     IDirect3DTexture9_GetSurfaceLevel(texture, 0, &surface);
     IDirect3DSurface9_GetDesc(surface, &desc);
-    expected = SUCCEEDED(IDirect3D9_CheckDeviceFormat(d3d, D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL,
-            D3DFMT_X8R8G8B8, 0, D3DRTYPE_TEXTURE, D3DFMT_A8L8)) ? D3DFMT_A8L8 : D3DFMT_A8R8G8B8;
-    ok(desc.Format == expected, "Returned format %u, expected %u.\n", desc.Format, expected);
+    ok(desc.Format == D3DFMT_A8L8, "Returned format %u, expected %u.\n", desc.Format, D3DFMT_A8L8);
     IDirect3DSurface9_Release(surface);
     IDirect3DTexture9_Release(texture);
     hr = D3DXCreateTextureFromFileInMemoryEx(device, png_grayscale, sizeof(png_grayscale),
@@ -2494,9 +2315,7 @@
     ok(hr == D3D_OK, "D3DXCreateTextureFromFileInMemoryEx returned %#lx, expected %#lx.\n", hr, D3D_OK);
     IDirect3DTexture9_GetSurfaceLevel(texture, 0, &surface);
     IDirect3DSurface9_GetDesc(surface, &desc);
-    expected = SUCCEEDED(IDirect3D9_CheckDeviceFormat(d3d, D3DADAPTER_DEFAULT, D3DDEVTYPE_HAL,
-            D3DFMT_X8R8G8B8, 0, D3DRTYPE_TEXTURE, D3DFMT_L8)) ? D3DFMT_L8 : D3DFMT_X8R8G8B8;
-    ok(desc.Format == expected, "Returned format %u, expected %u.\n", desc.Format, expected);
+    ok(desc.Format == D3DFMT_L8, "Returned format %u, expected %u.\n", desc.Format, D3DFMT_L8);
     IDirect3DSurface9_Release(surface);
     IDirect3DTexture9_Release(texture);
 
@@ -2691,7 +2510,6 @@
     check_texture_level_desc(texture, 0, D3DFMT_A16B16G16R16, D3DUSAGE_DYNAMIC, D3DPOOL_DEFAULT, 0, 0, 2, 2, FALSE);
     check_texture_level_desc(texture, 1, D3DFMT_A16B16G16R16, D3DUSAGE_DYNAMIC, D3DPOOL_DEFAULT, 0, 0, 1, 1, FALSE);
     IDirect3DTexture9_Release(texture);
-    IDirect3D9_Release(d3d);
 }
 
 static void test_D3DXCreateCubeTextureFromFileInMemory(IDirect3DDevice9 *device)
@@ -3272,37 +3090,6 @@
         out->x = 1;
 }
 
-static void set_vec2(D3DXVECTOR2 *v, float x, float y)
-{
-    v->x = x;
-    v->y = y;
-}
-
-static const D3DXVECTOR4 quadrant_color[] =
-{
-    { 1.0f, 0.0f, 0.0f, 1.0f }, { 0.0f, 1.0f, 0.0f, 1.0f }, { 0.0f, 0.0f, 1.0f, 1.0f }, { 1.0f, 1.0f, 1.0f, 1.0f },
-};
-
-static void WINAPI fill_func(D3DXVECTOR4 *value, const D3DXVECTOR2 *coord, const D3DXVECTOR2 *size, void *data)
-{
-    D3DXVECTOR2 vec = *coord;
-    unsigned int idx;
-
-    if (data)
-    {
-        *value = *(D3DXVECTOR4 *)data;
-        return;
-    }
-
-    set_vec2(&vec, (vec.x / size->x) - 0.5f, (vec.y / size->y) - 0.5f);
-    if (vec.x < 8.0f)
-        idx = vec.y < 8.0f ? 0 : 2;
-    else
-        idx = vec.y < 8.0f ? 1 : 3;
-
-    *value = quadrant_color[idx];
-}
-
 static void test_save_texture_to_dds_file(IDirect3DDevice9 *device)
 {
     static const struct
@@ -3327,91 +3114,74 @@
         uint32_t expected_caps;
         uint32_t expected_caps2;
         uint32_t expected_buffer_size;
-    } tests[] =
+        BOOL todo_hr;
+        BOOL todo_expected;
+    } dds_tests[] =
     {
         /* Paletted format tests. */
-        {
-            D3DRTYPE_TEXTURE, D3DFMT_P8, D3DPOOL_SCRATCH, 4, 4, 0, 3, test_palette, D3D_OK,
-            { 32, DDS_PF_INDEXED, 0, 8, 0, 0, 0, 0 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT), 4, 4, 0, 0, 3,
-            (DDSCAPS_TEXTURE | DDSCAPS_PALETTE | DDSCAPS_MIPMAP | DDSCAPS_COMPLEX),
-            0, PALETTED_DDS_FILE_HEADER_SIZE + 21,
+        { D3DRTYPE_TEXTURE, D3DFMT_P8, D3DPOOL_SCRATCH, 4, 4, 0, 3, test_palette, D3D_OK,
+          { 32, DDS_PF_INDEXED, 0, 8, 0, 0, 0, 0 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT), 4, 4, 0, 0, 3,
+          (DDSCAPS_TEXTURE | DDSCAPS_PALETTE | DDSCAPS_MIPMAP | DDSCAPS_COMPLEX), 0, PALETTED_DDS_FILE_HEADER_SIZE + 21,
         },
-        {
-            D3DRTYPE_CUBETEXTURE, D3DFMT_P8, D3DPOOL_SCRATCH, 4, 0, 0, 3, test_palette, D3D_OK,
-            { 32, DDS_PF_INDEXED, 0, 8, 0, 0, 0, 0 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT), 4, 4, 0, 0, 3,
-            (DDSCAPS_TEXTURE | DDSCAPS_PALETTE | DDSCAPS_MIPMAP | DDSCAPS_COMPLEX),
-            (DDSCAPS2_CUBEMAP | DDSCAPS2_CUBEMAP_ALLFACES), PALETTED_DDS_FILE_HEADER_SIZE + 126,
+        { D3DRTYPE_CUBETEXTURE, D3DFMT_P8, D3DPOOL_SCRATCH, 4, 0, 0, 3, test_palette, D3D_OK,
+          { 32, DDS_PF_INDEXED, 0, 8, 0, 0, 0, 0 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT), 4, 4, 0, 0, 3,
+          (DDSCAPS_TEXTURE | DDSCAPS_PALETTE | DDSCAPS_MIPMAP | DDSCAPS_COMPLEX), (DDSCAPS2_CUBEMAP | DDSCAPS2_CUBEMAP_ALLFACES),
+          PALETTED_DDS_FILE_HEADER_SIZE + (126)
         },
-        {
-            D3DRTYPE_VOLUMETEXTURE, D3DFMT_P8, D3DPOOL_SCRATCH, 4, 4, 4, 3, test_palette, D3D_OK,
-            { 32, DDS_PF_INDEXED, 0, 8, 0, 0, 0, 0 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT | DDSD_DEPTH), 4, 4, 0, 4, 3,
-            (DDSCAPS_TEXTURE | DDSCAPS_PALETTE | DDSCAPS_MIPMAP | DDSCAPS_COMPLEX),
-            DDSCAPS2_VOLUME, PALETTED_DDS_FILE_HEADER_SIZE + 73,
+        { D3DRTYPE_VOLUMETEXTURE, D3DFMT_P8, D3DPOOL_SCRATCH, 4, 4, 4, 3, test_palette, D3D_OK,
+          { 32, DDS_PF_INDEXED, 0, 8, 0, 0, 0, 0 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT | DDSD_DEPTH), 4, 4, 0, 4, 3,
+          (DDSCAPS_TEXTURE | DDSCAPS_PALETTE | DDSCAPS_MIPMAP | DDSCAPS_COMPLEX), (DDSCAPS2_VOLUME),
+          PALETTED_DDS_FILE_HEADER_SIZE + 73
         },
         /* D3DFMT_A8R8G8B8 textures with multiple levels. */
-        {
-            D3DRTYPE_TEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 4, 0, 3, NULL, D3D_OK,
-            { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT), 4, 4, 0, 0, 3,
-            (DDSCAPS_TEXTURE | DDSCAPS_MIPMAP | DDSCAPS_COMPLEX | DDSCAPS_ALPHA),
-            0, DDS_FILE_HEADER_SIZE + 84,
+        { D3DRTYPE_TEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 4, 0, 3, NULL, D3D_OK,
+          { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT), 4, 4, 0, 0, 3,
+          (DDSCAPS_TEXTURE | DDSCAPS_MIPMAP | DDSCAPS_COMPLEX | DDSCAPS_ALPHA), 0, DDS_FILE_HEADER_SIZE + 84,
         },
-        {
-            D3DRTYPE_CUBETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 0, 0, 3, NULL, D3D_OK,
-            { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT), 4, 4, 0, 0, 3,
-            (DDSCAPS_TEXTURE | DDSCAPS_MIPMAP | DDSCAPS_ALPHA | DDSCAPS_COMPLEX),
-            (DDSCAPS2_CUBEMAP | DDSCAPS2_CUBEMAP_ALLFACES), DDS_FILE_HEADER_SIZE + 504,
+        { D3DRTYPE_CUBETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 0, 0, 3, NULL, D3D_OK,
+          { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_MIPMAPCOUNT), 4, 4, 0, 0, 3,
+          (DDSCAPS_TEXTURE | DDSCAPS_MIPMAP | DDSCAPS_ALPHA | DDSCAPS_COMPLEX), (DDSCAPS2_CUBEMAP | DDSCAPS2_CUBEMAP_ALLFACES),
+          DDS_FILE_HEADER_SIZE + (504)
         },
-        /* 5 */
         /*
          * Volume texture with D3DPOOL default. Can't be mapped for read,
          * can't be saved.
          */
-        {
-            D3DRTYPE_VOLUMETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 4, 4, 3, NULL, D3DERR_INVALIDCALL,
-        },
+        { D3DRTYPE_VOLUMETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 4, 4, 3, NULL, D3DERR_INVALIDCALL },
         /* D3DPOOL_SYSTEMMEM can be saved. */
-        {
-            D3DRTYPE_VOLUMETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_SYSTEMMEM, 4, 4, 4, 3, NULL, D3D_OK,
-            { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_DEPTH | DDSD_MIPMAPCOUNT), 4, 4, 0, 4, 3,
-            (DDSCAPS_TEXTURE | DDSCAPS_MIPMAP | DDSCAPS_ALPHA | DDSCAPS_COMPLEX),
-            DDSCAPS2_VOLUME, DDS_FILE_HEADER_SIZE + 292,
+        { D3DRTYPE_VOLUMETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_SYSTEMMEM, 4, 4, 4, 3, NULL, D3D_OK,
+          { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_DEPTH | DDSD_MIPMAPCOUNT), 4, 4, 0, 4, 3,
+          (DDSCAPS_TEXTURE | DDSCAPS_MIPMAP | DDSCAPS_ALPHA | DDSCAPS_COMPLEX ), (DDSCAPS2_VOLUME), DDS_FILE_HEADER_SIZE + 292,
         },
         /* Single mip level, no mip flags. */
-        {
-            D3DRTYPE_TEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 4, 0, 1, NULL, D3D_OK,
-            { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT), 4, 4, 0, 0, 0,
-            (DDSCAPS_TEXTURE | DDSCAPS_ALPHA),
-            0, DDS_FILE_HEADER_SIZE + 64,
+        { D3DRTYPE_TEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 4, 0, 1, NULL, D3D_OK,
+          { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT), 4, 4, 0, 0, 0,
+          (DDSCAPS_TEXTURE | DDSCAPS_ALPHA), 0, DDS_FILE_HEADER_SIZE + 64,
         },
-        {
-            D3DRTYPE_CUBETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 0, 0, 1, NULL, D3D_OK,
-            { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT), 4, 4, 0, 0, 0,
-            (DDSCAPS_TEXTURE | DDSCAPS_ALPHA | DDSCAPS_COMPLEX),
-            (DDSCAPS2_CUBEMAP | DDSCAPS2_CUBEMAP_ALLFACES), DDS_FILE_HEADER_SIZE + 384,
+        /* 8. */
+        { D3DRTYPE_CUBETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_DEFAULT, 4, 0, 0, 1, NULL, D3D_OK,
+          { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT), 4, 4, 0, 0, 0,
+          (DDSCAPS_TEXTURE | DDSCAPS_ALPHA | DDSCAPS_COMPLEX), (DDSCAPS2_CUBEMAP | DDSCAPS2_CUBEMAP_ALLFACES),
+          DDS_FILE_HEADER_SIZE + (384)
         },
-        {
-            D3DRTYPE_VOLUMETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_SYSTEMMEM, 4, 4, 4, 1, NULL, D3D_OK,
-            { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_DEPTH), 4, 4, 0, 4, 0,
-            (DDSCAPS_TEXTURE | DDSCAPS_ALPHA),
-            DDSCAPS2_VOLUME, DDS_FILE_HEADER_SIZE + 256,
+        { D3DRTYPE_VOLUMETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_SYSTEMMEM, 4, 4, 4, 1, NULL, D3D_OK,
+          { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT | DDSD_DEPTH), 4, 4, 0, 4, 0,
+          (DDSCAPS_TEXTURE | DDSCAPS_ALPHA), (DDSCAPS2_VOLUME), DDS_FILE_HEADER_SIZE + 256,
         },
-        /* 10. */
         /* Volume texture with a depth of 1. */
-        {
-            D3DRTYPE_VOLUMETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_SYSTEMMEM, 4, 4, 1, 1, NULL, D3D_OK,
-            { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
-            (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT), 4, 4, 0, 0, 0,
-            (DDSCAPS_TEXTURE | DDSCAPS_ALPHA),
-            0, DDS_FILE_HEADER_SIZE + 64,
+        { D3DRTYPE_VOLUMETEXTURE, D3DFMT_A8R8G8B8, D3DPOOL_SYSTEMMEM, 4, 4, 1, 1, NULL, D3D_OK,
+          { 32, DDS_PF_RGB | DDS_PF_ALPHA, 0, 32, 0x00ff0000, 0x0000ff00, 0x000000ff, 0xff000000 },
+          (DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT), 4, 4, 0, 0, 0,
+          (DDSCAPS_TEXTURE | DDSCAPS_ALPHA), 0, DDS_FILE_HEADER_SIZE + 64,
         },
     };
     struct
@@ -3428,29 +3198,29 @@
     unsigned int i;
     HRESULT hr;
 
-    for (i = 0; i < ARRAY_SIZE(tests); ++i)
+    for (i = 0; i < ARRAY_SIZE(dds_tests); ++i)
     {
         winetest_push_context("Test %u", i);
 
-        switch (tests[i].type)
+        switch (dds_tests[i].type)
         {
             case D3DRTYPE_TEXTURE:
-                hr = IDirect3DDevice9_CreateTexture(device, tests[i].width, tests[i].height, tests[i].mip_levels, 0,
-                        tests[i].format, tests[i].pool, &texture, NULL);
+                hr = IDirect3DDevice9_CreateTexture(device, dds_tests[i].width, dds_tests[i].height, dds_tests[i].mip_levels, 0,
+                        dds_tests[i].format, dds_tests[i].pool, &texture, NULL);
                 ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
                 save_tex = (IDirect3DBaseTexture9 *)texture;
                 break;
 
             case D3DRTYPE_CUBETEXTURE:
-                hr = IDirect3DDevice9_CreateCubeTexture(device, tests[i].width, tests[i].mip_levels, 0,
-                        tests[i].format, tests[i].pool, &cube_texture, NULL);
+                hr = IDirect3DDevice9_CreateCubeTexture(device, dds_tests[i].width, dds_tests[i].mip_levels, 0,
+                        dds_tests[i].format, dds_tests[i].pool, &cube_texture, NULL);
                 ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
                 save_tex = (IDirect3DBaseTexture9 *)cube_texture;
                 break;
 
             case D3DRTYPE_VOLUMETEXTURE:
-                hr = IDirect3DDevice9_CreateVolumeTexture(device, tests[i].width, tests[i].height, tests[i].depth,
-                        tests[i].mip_levels, 0, tests[i].format, tests[i].pool, &volume_texture, NULL);
+                hr = IDirect3DDevice9_CreateVolumeTexture(device, dds_tests[i].width, dds_tests[i].height, dds_tests[i].depth,
+                        dds_tests[i].mip_levels, 0, dds_tests[i].format, dds_tests[i].pool, &volume_texture, NULL);
                 ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
                 save_tex = (IDirect3DBaseTexture9 *)volume_texture;
                 break;
@@ -3460,18 +3230,17 @@
                 break;
         }
 
-        hr = D3DXSaveTextureToFileInMemory(&buffer, D3DXIFF_DDS, save_tex, tests[i].palette);
-        ok(hr == tests[i].expected_hr, "Unexpected hr %#lx.\n", hr);
+        hr = D3DXSaveTextureToFileInMemory(&buffer, D3DXIFF_DDS, save_tex, dds_tests[i].palette);
+        todo_wine_if(dds_tests[i].todo_hr) ok(hr == dds_tests[i].expected_hr, "Unexpected hr %#lx.\n", hr);
         if (SUCCEEDED(hr))
         {
-            ok(ID3DXBuffer_GetBufferSize(buffer) == tests[i].expected_buffer_size, "Unexpected buffer size %lu.\n",
+            ok(ID3DXBuffer_GetBufferSize(buffer) == dds_tests[i].expected_buffer_size, "Unexpected buffer size %lu.\n",
                     ID3DXBuffer_GetBufferSize(buffer));
 
             dds = ID3DXBuffer_GetBufferPointer(buffer);
-            check_dds_header(&dds->header, tests[i].expected_flags, tests[i].expected_height, tests[i].expected_width,
-                    tests[i].expected_pitch, tests[i].expected_depth, tests[i].expected_mip_levels,
-                    &tests[i].expected_pixel_format, tests[i].expected_caps, tests[i].expected_caps2,
-                    FALSE);
+            check_dds_header(&dds->header, dds_tests[i].expected_flags, dds_tests[i].expected_height, dds_tests[i].expected_width,
+                    dds_tests[i].expected_pitch, dds_tests[i].expected_depth, dds_tests[i].expected_mip_levels,
+                    &dds_tests[i].expected_pixel_format, dds_tests[i].expected_caps, dds_tests[i].expected_caps2, dds_tests[i].todo_expected);
             ID3DXBuffer_Release(buffer);
         }
 
@@ -3482,9 +3251,6 @@
 
 static void test_D3DXSaveTextureToFileInMemory(IDirect3DDevice9 *device)
 {
-    const D3DXVECTOR4 clear_val = { 0.0f, 0.0f, 0.0f, 0.0f };
-    IDirect3DSurface9 *surface;
-    struct surface_readback rb;
     HRESULT hr;
     IDirect3DTexture9 *texture;
     IDirect3DCubeTexture9 *cube_texture;
@@ -3507,17 +3273,20 @@
     {
         hr = D3DXSaveTextureToFileInMemory(&buffer, file_format, (IDirect3DBaseTexture9 *)texture, NULL);
         ok(hr == D3D_OK, "D3DXSaveTextureToFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
-        buffer_pointer = ID3DXBuffer_GetBufferPointer(buffer);
-        buffer_size = ID3DXBuffer_GetBufferSize(buffer);
-        hr = D3DXGetImageInfoFromFileInMemory(buffer_pointer, buffer_size, &info);
-        ok(hr == D3D_OK, "D3DXGetImageInfoFromFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
-
-        ok(info.Width == 256, "Got width %u, expected %u\n", info.Width, 256);
-        ok(info.Height == 256, "Got height %u, expected %u\n", info.Height, 256);
-        ok(info.MipLevels == 1, "Got miplevels %u, expected %u\n", info.MipLevels, 1);
-        ok(info.ResourceType == D3DRTYPE_TEXTURE, "Got resource type %#x, expected %#x\n", info.ResourceType, D3DRTYPE_TEXTURE);
-        ok(info.ImageFileFormat == file_format, "Got file format %#x, expected %#x\n", info.ImageFileFormat, file_format);
-        ID3DXBuffer_Release(buffer);
+        if (SUCCEEDED(hr))
+        {
+            buffer_pointer = ID3DXBuffer_GetBufferPointer(buffer);
+            buffer_size = ID3DXBuffer_GetBufferSize(buffer);
+            hr = D3DXGetImageInfoFromFileInMemory(buffer_pointer, buffer_size, &info);
+            ok(hr == D3D_OK, "D3DXGetImageInfoFromFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
+
+            ok(info.Width == 256, "Got width %u, expected %u\n", info.Width, 256);
+            ok(info.Height == 256, "Got height %u, expected %u\n", info.Height, 256);
+            ok(info.MipLevels == 1, "Got miplevels %u, expected %u\n", info.MipLevels, 1);
+            ok(info.ResourceType == D3DRTYPE_TEXTURE, "Got resource type %#x, expected %#x\n", info.ResourceType, D3DRTYPE_TEXTURE);
+            ok(info.ImageFileFormat == file_format, "Got file format %#x, expected %#x\n", info.ImageFileFormat, file_format);
+            ID3DXBuffer_Release(buffer);
+        }
     }
 
     hr = D3DXSaveTextureToFileInMemory(&buffer, D3DXIFF_DDS, (IDirect3DBaseTexture9 *)texture, NULL);
@@ -3550,38 +3319,43 @@
 
     hr = D3DXSaveTextureToFileInMemory(&buffer, D3DXIFF_BMP, (IDirect3DBaseTexture9 *)cube_texture, NULL);
     ok(hr == D3D_OK, "D3DXSaveTextureToFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
-    buffer_pointer = ID3DXBuffer_GetBufferPointer(buffer);
-    buffer_size = ID3DXBuffer_GetBufferSize(buffer);
-    hr = D3DXGetImageInfoFromFileInMemory(buffer_pointer, buffer_size, &info);
-    ok(hr == D3D_OK, "D3DXGetImageInfoFromFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
-
-    ok(info.Width == 256, "Got width %u, expected %u\n", info.Width, 256);
-    ok(info.Height == 256, "Got height %u, expected %u\n", info.Height, 256);
-    ok(info.MipLevels == 1, "Got miplevels %u, expected %u\n", info.MipLevels, 1);
-    ok(info.ResourceType == D3DRTYPE_TEXTURE, "Got resource type %#x, expected %#x\n", info.ResourceType, D3DRTYPE_TEXTURE);
-    ok(info.ImageFileFormat == D3DXIFF_BMP, "Got file format %#x, expected %#x\n", info.ImageFileFormat, D3DXIFF_BMP);
-
-    /* positive x face is saved */
-    hr = IDirect3DDevice9_CreateOffscreenPlainSurface(device, 256, 256, D3DFMT_X8R8G8B8, D3DPOOL_SYSTEMMEM, &surface, NULL);
     if (SUCCEEDED(hr))
     {
-        D3DLOCKED_RECT locked_rect;
-
-        hr = D3DXLoadSurfaceFromFileInMemory(surface, NULL, NULL, buffer_pointer, buffer_size, NULL, D3DX_FILTER_NONE, 0, NULL);
-        ok(hr == D3D_OK, "D3DXLoadSurfaceFromFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
-
-        hr = IDirect3DSurface9_LockRect(surface, &locked_rect, NULL, D3DLOCK_READONLY);
+        IDirect3DSurface9 *surface;
+
+        buffer_pointer = ID3DXBuffer_GetBufferPointer(buffer);
+        buffer_size = ID3DXBuffer_GetBufferSize(buffer);
+        hr = D3DXGetImageInfoFromFileInMemory(buffer_pointer, buffer_size, &info);
+        ok(hr == D3D_OK, "D3DXGetImageInfoFromFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
+
+        ok(info.Width == 256, "Got width %u, expected %u\n", info.Width, 256);
+        ok(info.Height == 256, "Got height %u, expected %u\n", info.Height, 256);
+        ok(info.MipLevels == 1, "Got miplevels %u, expected %u\n", info.MipLevels, 1);
+        ok(info.ResourceType == D3DRTYPE_TEXTURE, "Got resource type %#x, expected %#x\n", info.ResourceType, D3DRTYPE_TEXTURE);
+        ok(info.ImageFileFormat == D3DXIFF_BMP, "Got file format %#x, expected %#x\n", info.ImageFileFormat, D3DXIFF_BMP);
+
+        /* positive x face is saved */
+        hr = IDirect3DDevice9_CreateOffscreenPlainSurface(device, 256, 256, D3DFMT_X8R8G8B8, D3DPOOL_SYSTEMMEM, &surface, NULL);
         if (SUCCEEDED(hr))
         {
-            DWORD *color = locked_rect.pBits;
-            ok(*color == 0x00ff0000, "Got color %#lx, expected %#x\n", *color, 0x00ff0000);
-            IDirect3DSurface9_UnlockRect(surface);
-        }
-
-        IDirect3DSurface9_Release(surface);
-    } else skip("Failed to create surface\n");
-
-    ID3DXBuffer_Release(buffer);
+            D3DLOCKED_RECT locked_rect;
+
+            hr = D3DXLoadSurfaceFromFileInMemory(surface, NULL, NULL, buffer_pointer, buffer_size, NULL, D3DX_FILTER_NONE, 0, NULL);
+            ok(hr == D3D_OK, "D3DXLoadSurfaceFromFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
+
+            hr = IDirect3DSurface9_LockRect(surface, &locked_rect, NULL, D3DLOCK_READONLY);
+            if (SUCCEEDED(hr))
+            {
+                DWORD *color = locked_rect.pBits;
+                ok(*color == 0x00ff0000, "Got color %#lx, expected %#x\n", *color, 0x00ff0000);
+                IDirect3DSurface9_UnlockRect(surface);
+            }
+
+            IDirect3DSurface9_Release(surface);
+        } else skip("Failed to create surface\n");
+
+        ID3DXBuffer_Release(buffer);
+    }
 
     hr = D3DXSaveTextureToFileInMemory(&buffer, D3DXIFF_DDS, (IDirect3DBaseTexture9 *)cube_texture, NULL);
     ok(hr == D3D_OK, "D3DXSaveTextureToFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
@@ -3610,18 +3384,21 @@
 
     hr = D3DXSaveTextureToFileInMemory(&buffer, D3DXIFF_BMP, (IDirect3DBaseTexture9 *)volume_texture, NULL);
     ok(hr == D3D_OK, "D3DXSaveTextureToFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
-    buffer_pointer = ID3DXBuffer_GetBufferPointer(buffer);
-    buffer_size = ID3DXBuffer_GetBufferSize(buffer);
-    hr = D3DXGetImageInfoFromFileInMemory(buffer_pointer, buffer_size, &info);
-    ok(hr == D3D_OK, "D3DXGetImageInfoFromFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
-
-    ok(info.Width == 256, "Got width %u, expected %u\n", info.Width, 256);
-    ok(info.Height == 256, "Got height %u, expected %u\n", info.Height, 256);
-    ok(info.Depth == 1, "Got depth %u, expected %u\n", info.Depth, 1);
-    ok(info.MipLevels == 1, "Got miplevels %u, expected %u\n", info.MipLevels, 1);
-    ok(info.ResourceType == D3DRTYPE_TEXTURE, "Got resource type %#x, expected %#x\n", info.ResourceType, D3DRTYPE_TEXTURE);
-    ok(info.ImageFileFormat == D3DXIFF_BMP, "Got file format %#x, expected %#x\n", info.ImageFileFormat, D3DXIFF_BMP);
-    ID3DXBuffer_Release(buffer);
+    if (SUCCEEDED(hr))
+    {
+        buffer_pointer = ID3DXBuffer_GetBufferPointer(buffer);
+        buffer_size = ID3DXBuffer_GetBufferSize(buffer);
+        hr = D3DXGetImageInfoFromFileInMemory(buffer_pointer, buffer_size, &info);
+        ok(hr == D3D_OK, "D3DXGetImageInfoFromFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
+
+        ok(info.Width == 256, "Got width %u, expected %u\n", info.Width, 256);
+        ok(info.Height == 256, "Got height %u, expected %u\n", info.Height, 256);
+        ok(info.Depth == 1, "Got depth %u, expected %u\n", info.Depth, 1);
+        ok(info.MipLevels == 1, "Got miplevels %u, expected %u\n", info.MipLevels, 1);
+        ok(info.ResourceType == D3DRTYPE_TEXTURE, "Got resource type %#x, expected %#x\n", info.ResourceType, D3DRTYPE_TEXTURE);
+        ok(info.ImageFileFormat == D3DXIFF_BMP, "Got file format %#x, expected %#x\n", info.ImageFileFormat, D3DXIFF_BMP);
+        ID3DXBuffer_Release(buffer);
+    }
 
     hr = D3DXSaveTextureToFileInMemory(&buffer, D3DXIFF_DDS, (IDirect3DBaseTexture9 *)volume_texture, NULL);
     ok(hr == D3D_OK, "D3DXSaveTextureToFileInMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
@@ -3640,58 +3417,6 @@
     ID3DXBuffer_Release(buffer);
 
     IDirect3DVolumeTexture9_Release(volume_texture);
-
-    /* Test JPEG compression quality. */
-    hr = IDirect3DDevice9_CreateTexture(device, 16, 16, 1, 0, D3DFMT_X8R8G8B8, D3DPOOL_MANAGED, &texture, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = D3DXFillTexture(texture, fill_func, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = D3DXSaveTextureToFileInMemory(&buffer, D3DXIFF_JPG, (IDirect3DBaseTexture9 *)texture, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = D3DXFillTexture(texture, fill_func, (void *)&clear_val);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IDirect3DTexture9_GetSurfaceLevel(texture, 0, &surface);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = D3DXLoadSurfaceFromFileInMemory(surface, NULL, NULL, ID3DXBuffer_GetBufferPointer(buffer),
-            ID3DXBuffer_GetBufferSize(buffer), NULL, D3DX_FILTER_NONE, 0, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    IDirect3DSurface9_Release(surface);
-    ID3DXBuffer_Release(buffer);
-
-    /*
-     * Wine's JPEG compression quality is worse than native. Native uses 4:2:0
-     * subsampling which is the same as what we use, but whatever compression
-     * settings they're using results in a JPEG image that is much closer to
-     * the original uncompressed surface. Native compresses with a quality
-     * setting of 95 while currently we use the libjpeg default of 75.
-     * However, even if we up our quality to 95 to match we still end up with
-     * a max_diff of 14, so there must also be some other setting(s) involved.
-     * The quality difference is most apparent in 16x16 blocks with multiple
-     * colors.
-     */
-    get_texture_surface_readback(device, texture, 0, &rb);
-    todo_wine ok(compare_color(get_readback_color(&rb,  0,  0),  0xffff0000, 5),
-            "Unexpected color %#x, max_diff %d.\n", get_readback_color(&rb, 0, 0),
-            pixel_4bpp_max_channel_diff(get_readback_color(&rb, 0, 0), 0xffff0000));
-    todo_wine ok(compare_color(get_readback_color(&rb, 15,  0),  0xff00ff00, 3),
-            "Unexpected color %#x, max_diff %d.\n", get_readback_color(&rb, 15, 0),
-            pixel_4bpp_max_channel_diff(get_readback_color(&rb, 15, 0), 0xff00ff00));
-    todo_wine ok(compare_color(get_readback_color(&rb,  0, 15),  0xff0000ff, 6),
-            "Unexpected color %#x, max_diff %d.\n", get_readback_color(&rb, 0, 15),
-            pixel_4bpp_max_channel_diff(get_readback_color(&rb, 0, 15), 0xff0000ff));
-    todo_wine ok(compare_color(get_readback_color(&rb, 15, 15),  0xffffffff, 1),
-            "Unexpected color %#x, max_diff %d.\n", get_readback_color(&rb, 15, 15),
-            pixel_4bpp_max_channel_diff(get_readback_color(&rb, 15, 15), 0xffffffff));
-    release_surface_readback(&rb);
-
-    IDirect3DTexture9_Release(texture);
-
     test_save_texture_to_dds_file(device);
 }
 
