diff --git a/dlls/kernel32/tests/actctx.c b/dlls/kernel32/tests/actctx.c
--- a/dlls/kernel32/tests/actctx.c
+++ b/dlls/kernel32/tests/actctx.c
@@ -28,47 +28,12 @@
 
 #include "oaidl.h"
 #include "initguid.h"
-#include "wine/asm.h"
-
-#ifdef __ASM_USE_FASTCALL_WRAPPER
-
-extern void * WINAPI wrap_fastcall_func1( void *func, const void *a );
-__ASM_STDCALL_FUNC( wrap_fastcall_func1, 8,
-                   "popl %ecx\n\t"
-                   "popl %eax\n\t"
-                   "xchgl (%esp),%ecx\n\t"
-                   "jmp *%eax" );
-extern void * WINAPI wrap_fastcall_func2( void *func, const void *a, const void *b );
-__ASM_STDCALL_FUNC( wrap_fastcall_func2, 12,
-                   "popl %edx\n\t"
-                   "popl %eax\n\t"
-                   "popl %ecx\n\t"
-                   "xchgl (%esp),%edx\n\t"
-                   "jmp *%eax" );
-
-#define call_fastcall_func1(func,a) wrap_fastcall_func1(func,a)
-#define call_fastcall_func2(func,a,b) wrap_fastcall_func2(func,a,b)
-
-#else  /* __ASM_USE_FASTCALL_WRAPPER */
-
-#define call_fastcall_func1(func,a) func(a)
-#define call_fastcall_func2(func,a,b) func(a,b)
-
-#endif  /* __ASM_USE_FASTCALL_WRAPPER */
+
 static BOOL   (WINAPI *pQueryActCtxSettingsW)(DWORD,HANDLE,LPCWSTR,LPCWSTR,LPWSTR,SIZE_T,SIZE_T*);
 
-static NTSTATUS (NTAPI *pRtlActivateActivationContext)(ULONG,struct _ACTIVATION_CONTEXT *,ULONG_PTR *);
-static NTSTATUS (NTAPI *pRtlActivateActivationContextEx)(ULONG,TEB *,struct _ACTIVATION_CONTEXT *,ULONG_PTR *);
-static PRTL_ACTIVATION_CONTEXT_STACK_FRAME (FASTCALL *pRtlActivateActivationContextUnsafeFast)(PRTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_EXTENDED,
-                                                                                               struct _ACTIVATION_CONTEXT *);
-static VOID (FASTCALL *pRtlDeactivateActivationContextUnsafeFast)(PRTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_EXTENDED);
 static NTSTATUS(NTAPI *pRtlFindActivationContextSectionString)(DWORD,const GUID *,ULONG,PUNICODE_STRING,PACTCTX_SECTION_KEYED_DATA);
-static VOID (NTAPI *pRtlFreeThreadActivationContextStack)(VOID);
 static BOOLEAN (NTAPI *pRtlCreateUnicodeStringFromAsciiz)(PUNICODE_STRING, PCSZ);
 static VOID    (NTAPI *pRtlFreeUnicodeString)(PUNICODE_STRING);
-static NTSTATUS(NTAPI *pRtlQueryInformationActiveActivationContext)(ULONG,PVOID,SIZE_T,SIZE_T *);
-
-#define NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED 0x8 /* RTL_ACTIVATION_CONTEXT_STACK_FRAME.Flags */
 
 #ifdef __i386__
 #define ARCH "x86"
@@ -569,26 +534,11 @@
 "  <file name=\"sxs_dll.dll\"></file>"
 "</assembly>";
 
-static const char two_dll_manifest_dll_fr[] =
-"<assembly xmlns=\"urn:schemas-microsoft-com:asm.v3\" manifestVersion=\"1.0\">"
-"  <assemblyIdentity type=\"win32\" name=\"sxs_dll\" version=\"1.0.0.0\" processorArchitecture=\"x86\" publicKeyToken=\"0000000000000000\" language=\"fr-FR\"/>"
-"  <file name=\"sxs_dll.dll\"></file>"
-"</assembly>";
-
 static const char two_dll_manifest_exe[] =
 "<assembly xmlns=\"urn:schemas-microsoft-com:asm.v1\" manifestVersion=\"1.0\">"
 "  <dependency>"
 "    <dependentAssembly>"
 "      <assemblyIdentity type=\"win32\" name=\"sxs_dll\" version=\"1.0.0.0\" processorArchitecture=\"x86\" publicKeyToken=\"0000000000000000\" language=\"*\"/>"
-"    </dependentAssembly>"
-"  </dependency>"
-"</assembly>";
-
-static const char two_dll_manifest_exe_fr[] =
-"<assembly xmlns=\"urn:schemas-microsoft-com:asm.v1\" manifestVersion=\"1.0\">"
-"  <dependency>"
-"    <dependentAssembly>"
-"      <assemblyIdentity type=\"win32\" name=\"sxs_dll\" version=\"1.0.0.0\" processorArchitecture=\"x86\" publicKeyToken=\"0000000000000000\" language=\"fr-FR\"/>"
 "    </dependentAssembly>"
 "  </dependency>"
 "</assembly>";
@@ -2231,9 +2181,7 @@
 
 static void test_actctx(void)
 {
-    RTL_ACTIVATION_CONTEXT_STACK_FRAME *frame;
     ULONG_PTR cookie;
-    NTSTATUS status;
     HANDLE handle;
     BOOL b;
 
@@ -2246,62 +2194,6 @@
         test_basic_info(handle, __LINE__);
         test_detailed_info(handle, &detailed_info0, __LINE__);
         test_runlevel_info(handle, &runlevel_info0, __LINE__);
-        ReleaseActCtx(handle);
-    }
-
-    /* Test flags for normal frames */
-    if (!create_manifest_file("test1.manifest", manifest1, -1, NULL, NULL))
-    {
-        skip("Could not create manifest file\n");
-        return;
-    }
-    handle = test_create("test1.manifest");
-    ok(handle != INVALID_HANDLE_VALUE, "handle == INVALID_HANDLE_VALUE, error %lu\n", GetLastError());
-    DeleteFileA("test1.manifest");
-    if (handle != INVALID_HANDLE_VALUE)
-    {
-        b = ActivateActCtx(handle, &cookie);
-        ok(b, "ActivateActCtx failed: %lu\n", GetLastError());
-
-        frame = NtCurrentTeb()->ActivationContextStackPointer->ActiveFrame;
-        ok(!frame->Previous, "Got unexpected Previous.\n");
-        ok(frame->ActivationContext == handle, "Got unexpected ActivationContext.\n");
-        ok(frame->Flags & NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED, "Got unexpected Flags %#lx.\n", frame->Flags);
-
-        b = DeactivateActCtx(0, cookie);
-        ok(b, "DeactivateActCtx failed: %lu\n", GetLastError());
-        b = GetCurrentActCtx(&handle);
-        ok(handle == NULL, "handle = %p, expected NULL\n", handle);
-        ok(b, "GetCurrentActCtx failed: %lu\n", GetLastError());
-
-        status = pRtlActivateActivationContext(0, handle, &cookie);
-        ok(status == STATUS_SUCCESS,  "Got unexpected status %#lx.\n", status);
-
-        frame = NtCurrentTeb()->ActivationContextStackPointer->ActiveFrame;
-        ok(!frame->Previous, "Got unexpected Previous.\n");
-        ok(frame->ActivationContext == handle, "Got unexpected ActivationContext.\n");
-        ok(frame->Flags & NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED, "Got unexpected Flags %#lx.\n", frame->Flags);
-
-        b = DeactivateActCtx(0, cookie);
-        ok(b, "DeactivateActCtx failed: %lu\n", GetLastError());
-        b = GetCurrentActCtx(&handle);
-        ok(handle == NULL, "handle = %p, expected NULL\n", handle);
-        ok(b, "GetCurrentActCtx failed: %lu\n", GetLastError());
-
-        status = pRtlActivateActivationContextEx(0, NtCurrentTeb(), handle, &cookie);
-        ok(status == STATUS_SUCCESS,  "Got unexpected status %#lx.\n", status);
-
-        frame = NtCurrentTeb()->ActivationContextStackPointer->ActiveFrame;
-        ok(!frame->Previous, "Got unexpected Previous.\n");
-        ok(frame->ActivationContext == handle, "Got unexpected ActivationContext.\n");
-        ok(frame->Flags & NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED, "Got unexpected Flags %#lx.\n", frame->Flags);
-
-        b = DeactivateActCtx(0, cookie);
-        ok(b, "DeactivateActCtx failed: %lu\n", GetLastError());
-        b = GetCurrentActCtx(&handle);
-        ok(handle == NULL, "handle = %p, expected NULL\n", handle);
-        ok(b, "GetCurrentActCtx failed: %lu\n", GetLastError());
-
         ReleaseActCtx(handle);
     }
 
@@ -2425,7 +2317,10 @@
     }
 
     handle = test_create("test3.manifest");
-    ok(handle != INVALID_HANDLE_VALUE, "handle == INVALID_HANDLE_VALUE, error %lu\n", GetLastError());
+    ok(handle != INVALID_HANDLE_VALUE || broken(handle == INVALID_HANDLE_VALUE) /* XP pre-SP2, win2k3 w/o SP */,
+        "handle == INVALID_HANDLE_VALUE, error %lu\n", GetLastError());
+    if (handle == INVALID_HANDLE_VALUE)
+        win_skip("Some activation context features not supported, skipping a test (possibly old XP/Win2k3 system\n");
     DeleteFileA("test3.manifest");
     if(handle != INVALID_HANDLE_VALUE) {
         static const WCHAR nameW[] = {'t','e','s','t','s','u','r','r','o','g','a','t','e',0};
@@ -2478,11 +2373,15 @@
 
     if(create_manifest_file("test6.manifest", manifest6, -1, NULL, NULL)) {
         handle = test_create("test6.manifest");
-        ok(handle != INVALID_HANDLE_VALUE, "Unexpected context handle %p.\n", handle);
+        ok(handle != INVALID_HANDLE_VALUE || broken(handle == INVALID_HANDLE_VALUE) /* WinXP */,
+            "Unexpected context handle %p.\n", handle);
         DeleteFileA("test6.manifest");
         DeleteFileA("testdep.manifest");
-        test_runlevel_info(handle, &runlevel_info6, __LINE__);
-        ReleaseActCtx(handle);
+        if(handle != INVALID_HANDLE_VALUE)
+        {
+            test_runlevel_info(handle, &runlevel_info6, __LINE__);
+            ReleaseActCtx(handle);
+        }
     }
     else
         skip("Could not create manifest file 6\n");
@@ -2825,7 +2724,9 @@
     sprintf(cmdline, "\"%s\" %s manifest1", argv[0], argv[1]);
     ret = CreateProcessA(argv[0], cmdline, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
     ok(ret, "Could not create process: %lu\n", GetLastError());
-    wait_child_process( &pi );
+    wait_child_process( pi.hProcess );
+    CloseHandle(pi.hThread);
+    CloseHandle(pi.hProcess);
     DeleteFileA(path);
 }
 
@@ -3088,7 +2989,9 @@
     SetLastError(0xdeadbeef);
     handle = CreateActCtxA(&actctx);
     ok(handle == INVALID_HANDLE_VALUE, "got handle %p\n", handle);
-    ok(GetLastError()==ERROR_PATH_NOT_FOUND, "got error %ld\n", GetLastError());
+    ok(GetLastError()==ERROR_PATH_NOT_FOUND ||
+            broken(GetLastError()==ERROR_FILE_NOT_FOUND) /* WinXP */,
+            "got error %ld\n", GetLastError());
 
     CreateDirectoryA(dir, NULL);
     memset(&actctx, 0, sizeof(actctx));
@@ -3186,15 +3089,9 @@
     pQueryActCtxSettingsW = (void *)GetProcAddress( hLibrary, "QueryActCtxSettingsW" );
 
     hLibrary = GetModuleHandleA("ntdll.dll");
-    X(RtlActivateActivationContext);
-    X(RtlActivateActivationContextEx);
-    X(RtlActivateActivationContextUnsafeFast);
-    X(RtlDeactivateActivationContextUnsafeFast);
     X(RtlFindActivationContextSectionString);
-    X(RtlFreeThreadActivationContextStack);
     X(RtlCreateUnicodeStringFromAsciiz);
     X(RtlFreeUnicodeString);
-    X(RtlQueryInformationActiveActivationContext);
 #undef X
 
     return TRUE;
@@ -3648,7 +3545,6 @@
 typedef struct
 {
     char path_tmp[MAX_PATH];
-    char path_lang[MAX_PATH];
     char path_dll[MAX_PATH + 11];
     char path_manifest_exe[MAX_PATH + 12];
     char path_manifest_dll[MAX_PATH + 16];
@@ -3659,46 +3555,42 @@
     void (WINAPI *get_path)(char *buffer, int buffer_size);
 } sxs_info;
 
-static void fill_sxs_paths(sxs_info *info, const char *temp, const char *path_dll, const char *exe_manifest,
-                           const char *dll_manifest, const char *lang)
-{
+static BOOL fill_sxs_info(sxs_info *info, const char *temp, const char *path_dll, const char *exe_manifest, const char *dll_manifest, BOOL do_load)
+{
+    BOOL success;
+
+    GetTempPathA(MAX_PATH, info->path_tmp);
     strcat(info->path_tmp, temp);
     strcat(info->path_tmp, "\\");
     CreateDirectoryA(info->path_tmp, NULL);
 
-    if (lang)
-    {
-        sprintf( info->path_lang, "%s%s\\", info->path_tmp, lang );
-        CreateDirectoryA(info->path_lang, NULL);
-    }
-    else info->path_lang[0] = 0;
-
-    sprintf(info->path_dll, "%s%s", lang ? info->path_lang : info->path_tmp, "sxs_dll.dll");
+    sprintf(info->path_dll, "%s%s", info->path_tmp, "sxs_dll.dll");
     extract_resource(path_dll, "TESTDLL", info->path_dll);
 
     sprintf(info->path_manifest_exe, "%s%s", info->path_tmp, "exe.manifest");
     create_manifest_file(info->path_manifest_exe, exe_manifest, -1, NULL, NULL);
 
-    sprintf(info->path_manifest_dll, "%s%s", lang ? info->path_lang : info->path_tmp, "sxs_dll.manifest");
+    sprintf(info->path_manifest_dll, "%s%s", info->path_tmp, "sxs_dll.manifest");
     create_manifest_file(info->path_manifest_dll, dll_manifest, -1, NULL, NULL);
-}
-
-static void fill_sxs_info(sxs_info *info, const char *temp, const char *path_dll, const char *exe_manifest,
-                          const char *dll_manifest, BOOL do_load)
-{
-    GetTempPathA(MAX_PATH, info->path_tmp);
-    fill_sxs_paths( info, temp, path_dll, exe_manifest, dll_manifest, NULL );
+
     info->context.cbSize = sizeof(ACTCTXA);
     info->context.lpSource = info->path_manifest_exe;
     info->context.lpAssemblyDirectory = info->path_tmp;
     info->context.dwFlags = ACTCTX_FLAG_ASSEMBLY_DIRECTORY_VALID;
 
     info->handle_context = CreateActCtxA(&info->context);
-    ok(info->handle_context != INVALID_HANDLE_VALUE, "CreateActCtxA failed: %ld\n", GetLastError());
+    ok((info->handle_context != NULL && info->handle_context != INVALID_HANDLE_VALUE )
+            || broken(GetLastError() == ERROR_SXS_CANT_GEN_ACTCTX), /* XP doesn't support manifests outside of PE files */
+            "CreateActCtxA failed: %ld\n", GetLastError());
+    if (GetLastError() == ERROR_SXS_CANT_GEN_ACTCTX)
+    {
+        skip("Failed to create activation context.\n");
+        return FALSE;
+    }
 
     if (do_load)
     {
-        BOOL success = ActivateActCtx(info->handle_context, &info->cookie);
+        success = ActivateActCtx(info->handle_context, &info->cookie);
         ok(success, "ActivateActCtx failed: %ld\n", GetLastError());
 
         info->module = LoadLibraryA("sxs_dll.dll");
@@ -3709,6 +3601,7 @@
 
         DeactivateActCtx(0, info->cookie);
     }
+    return TRUE;
 }
 
 static void clean_sxs_info(sxs_info *info)
@@ -3729,11 +3622,6 @@
     {
         BOOL ret = DeleteFileA(info->path_manifest_dll);
         ok(ret, "DeleteFileA failed for %s: %ld\n", info->path_manifest_dll, GetLastError());
-    }
-    if (*info->path_lang)
-    {
-        BOOL ret = RemoveDirectoryA(info->path_lang);
-        ok(ret, "RemoveDirectoryA failed for %s: %ld\n", info->path_lang, GetLastError());
     }
     if (*info->path_tmp)
     {
@@ -3757,8 +3645,10 @@
     sxs_info dll_2;
     char path1[MAX_PATH], path2[MAX_PATH];
 
-    fill_sxs_info(&dll_1, "1", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, TRUE);
-    fill_sxs_info(&dll_2, "2", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, TRUE);
+    if (!fill_sxs_info(&dll_1, "1", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, TRUE))
+        goto cleanup1;
+    if (!fill_sxs_info(&dll_2, "2", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, TRUE))
+        goto cleanup2;
 
     ok(dll_1.module != dll_2.module, "Libraries are the same\n");
     dll_1.get_path(path1, sizeof(path1));
@@ -3766,9 +3656,11 @@
     dll_2.get_path(path2, sizeof(path2));
     ok(strcmp(path2, dll_2.path_dll) == 0, "Got '%s', expected '%s'\n", path2, dll_2.path_dll);
 
+cleanup2:
     if (dll_2.module)
         FreeLibrary(dll_2.module);
     clean_sxs_info(&dll_2);
+cleanup1:
     if (dll_1.module)
         FreeLibrary(dll_1.module);
     clean_sxs_info(&dll_1);
@@ -3792,13 +3684,16 @@
     module = LoadLibraryA(path_dll_local);
     get_path = (void *)GetProcAddress(module, "get_path");
 
-    fill_sxs_info(&dll, "1", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, TRUE);
+    if (!fill_sxs_info(&dll, "1", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, TRUE))
+        goto cleanup;
+
     ok(dll.module != module, "Libraries are the same\n");
     dll.get_path(path1, sizeof(path1));
     ok(strcmp(path1, dll.path_dll) == 0, "Got '%s', expected '%s'\n", path1, dll.path_dll);
     get_path(path2, sizeof(path2));
     ok(strcmp(path2, path_dll_local) == 0, "Got '%s', expected '%s'\n", path2, path_dll_local);
 
+cleanup:
     if (module)
         FreeLibrary(module);
     if (dll.module)
@@ -3826,7 +3721,9 @@
     sprintf(path_dll_local, "%s%s", path_application, "sxs_dll.dll");
     extract_resource("dummy.dll", "TESTDLL", path_dll_local);
 
-    fill_sxs_info(&dll, "1", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, TRUE);
+    if (!fill_sxs_info(&dll, "1", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, TRUE))
+        goto cleanup;
+
     module = LoadLibraryA(path_dll_local);
     get_path = (void *)GetProcAddress(module, "get_path");
 
@@ -3836,6 +3733,7 @@
     get_path(path2, sizeof(path2));
     ok(strcmp(path2, path_dll_local) == 0, "Got '%s', expected '%s'\n", path2, path_dll_local);
 
+cleanup:
     if (module)
         FreeLibrary(module);
     if (dll.module)
@@ -3848,31 +3746,8 @@
     clean_sxs_info(&dll);
 }
 
-#define check_redirected_flag(a, b) _check_redirected_flag(__LINE__, a, b)
-static void _check_redirected_flag(int line, HMODULE module, BOOL redirected)
-{
-    LIST_ENTRY *root = &NtCurrentTeb()->Peb->LdrData->InLoadOrderModuleList;
-    BOOL found_dll = FALSE;
-
-    for (LIST_ENTRY *entry = root->Flink; entry != root; entry = entry->Flink)
-    {
-        LDR_DATA_TABLE_ENTRY *mod = CONTAINING_RECORD(entry, LDR_DATA_TABLE_ENTRY, InLoadOrderLinks);
-
-        if (mod->DllBase != module)
-            continue;
-
-        if (redirected)
-            ok_(__FILE__, line)(mod->Flags & LDR_REDIRECTED, "Got unexpected flags %#lx.\n", mod->Flags);
-        else
-            ok_(__FILE__, line)(!(mod->Flags & LDR_REDIRECTED), "Got unexpected flags %#lx.\n", mod->Flags);
-
-        found_dll = TRUE;
-        break;
-    }
-    ok(found_dll, "Couldn't find module %p in module list.\n", module);
-}
-
-/* Test GetModuleHandleA() with DLL base names in different context states */
+
+/* Test if we can get a module handle from loaded normal dll while context is active */
 static void test_one_with_sxs_and_GetModuleHandleA(void)
 {
     sxs_info dll;
@@ -3887,38 +3762,21 @@
     extract_resource("dummy.dll", "TESTDLL", path_dll_local);
 
     module = LoadLibraryA(path_dll_local);
-    check_redirected_flag(module, FALSE);
-
-    fill_sxs_info(&dll, "1", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, FALSE);
+
+    if (!fill_sxs_info(&dll, "1", "dummy.dll", two_dll_manifest_exe, two_dll_manifest_dll, FALSE))
+       goto cleanup;
+
     success = ActivateActCtx(dll.handle_context, &dll.cookie);
     ok(success, "ActivateActCtx failed: %ld\n", GetLastError());
 
-    /* Loaded normal dll can't be found while context is active */
     module_temp = GetModuleHandleA("sxs_dll.dll");
     ok (module_temp == 0, "Expected 0, got %p\n", module_temp);
 
-    dll.module = LoadLibraryA("sxs_dll.dll");
-    ok(dll.module != NULL && dll.module != module, "LoadLibrary failed\n");
-    check_redirected_flag(dll.module, TRUE);
-
-    /* Loaded SxS dll should be found while context is active */
-    module_temp = GetModuleHandleA("sxs_dll.dll");
-    ok(module_temp == dll.module, "Got unexpected module.\n");
-
     DeactivateActCtx(0, dll.cookie);
-    check_redirected_flag(dll.module, TRUE);
-
-    /* Normal loaded dll should be found while context is inactive */
-    module_temp = GetModuleHandleA("sxs_dll.dll");
-    ok(module_temp == module, "Got unexpected module.\n");
-
+
+cleanup:
     if (module)
         FreeLibrary(module);
-
-    /* Loaded SxS dll can't be found while context is inactive */
-    module_temp = GetModuleHandleA("sxs_dll.dll");
-    ok(module_temp == NULL, "Got unexpected module.\n");
-
     if (dll.module)
         FreeLibrary(dll.module);
     if (*path_dll_local)
@@ -3952,7 +3810,9 @@
     context.dwFlags = ACTCTX_FLAG_ASSEMBLY_DIRECTORY_VALID;
 
     handle_context = CreateActCtxA(&context);
-    ok(handle_context != INVALID_HANDLE_VALUE, "CreateActCtxA failed: %ld\n", GetLastError());
+    ok((handle_context != NULL && handle_context != INVALID_HANDLE_VALUE )
+        || broken(GetLastError() == ERROR_SXS_CANT_GEN_ACTCTX), /* XP doesn't support manifests outside of PE files */
+        "CreateActCtxA failed: %ld\n", GetLastError());
     if (GetLastError() == ERROR_SXS_CANT_GEN_ACTCTX)
     {
         skip("Failed to create activation context.\n");
@@ -3990,99 +3850,6 @@
         success = DeleteFileA(path_manifest);
         ok(success, "DeleteFileA failed for %s: %ld\n", path_manifest, GetLastError());
     }
-}
-
-static void test_manifest_lang(void)
-{
-    sxs_info dll;
-    char *p;
-
-    /* tmp path without language prefix */
-
-    GetTempPathA(MAX_PATH, dll.path_tmp);
-    fill_sxs_paths( &dll, "1", "dummy.dll", two_dll_manifest_exe_fr, two_dll_manifest_dll_fr, NULL );
-
-    dll.context.cbSize = sizeof(ACTCTXA);
-    dll.context.lpSource = dll.path_manifest_exe;
-    dll.context.lpAssemblyDirectory = dll.path_tmp;
-    dll.context.dwFlags = ACTCTX_FLAG_ASSEMBLY_DIRECTORY_VALID;
-
-    dll.handle_context = CreateActCtxA(&dll.context);
-    ok( dll.handle_context == INVALID_HANDLE_VALUE, "CreateActCtxA succeeded\n" );
-    ok( GetLastError() == ERROR_SXS_CANT_GEN_ACTCTX, "wrong error %ld\n", GetLastError() );
-
-    clean_sxs_info( &dll );
-
-    /* tmp path with language prefix */
-
-    GetTempPathA(MAX_PATH, dll.path_tmp);
-    fill_sxs_paths( &dll, "1", "dummy.dll", two_dll_manifest_exe_fr, two_dll_manifest_dll_fr, "fr-FR" );
-
-    dll.context.cbSize = sizeof(ACTCTXA);
-    dll.context.lpSource = dll.path_manifest_exe;
-    dll.context.lpAssemblyDirectory = dll.path_tmp;
-    dll.context.dwFlags = ACTCTX_FLAG_ASSEMBLY_DIRECTORY_VALID;
-
-    dll.handle_context = CreateActCtxA(&dll.context);
-    ok( dll.handle_context != INVALID_HANDLE_VALUE, "CreateActCtxA failed: %ld\n", GetLastError() );
-
-    clean_sxs_info( &dll );
-
-    /* exe path without language prefix */
-
-    GetModuleFileNameA(0, dll.path_tmp, MAX_PATH);
-    if ((p = strrchr( dll.path_tmp, '\\' ))) p[1] = 0;
-    dll.path_lang[0] = 0;
-
-    sprintf(dll.path_manifest_exe, "%s%s", dll.path_tmp, "exe.manifest");
-    create_manifest_file(dll.path_manifest_exe, two_dll_manifest_exe_fr, -1, NULL, NULL);
-
-    strcat(dll.path_tmp, "sxs_dll\\");
-    CreateDirectoryA(dll.path_tmp, NULL);
-    sprintf(dll.path_dll, "%s%s", dll.path_tmp, "sxs_dll.dll");
-    extract_resource("dummy.dll", "TESTDLL", dll.path_dll);
-
-    sprintf(dll.path_manifest_dll, "%s%s", dll.path_tmp, "sxs_dll.manifest");
-    create_manifest_file(dll.path_manifest_dll, two_dll_manifest_dll_fr, -1, NULL, NULL);
-
-    dll.context.cbSize = sizeof(ACTCTXA);
-    dll.context.lpSource = dll.path_manifest_exe;
-    dll.context.lpAssemblyDirectory = dll.path_tmp;
-    dll.context.dwFlags = ACTCTX_FLAG_ASSEMBLY_DIRECTORY_VALID;
-
-    dll.handle_context = CreateActCtxA(&dll.context);
-    ok( dll.handle_context == INVALID_HANDLE_VALUE, "CreateActCtxA succeeded\n" );
-    ok( GetLastError() == ERROR_SXS_CANT_GEN_ACTCTX, "wrong error %ld\n", GetLastError() );
-
-    clean_sxs_info( &dll );
-
-    /* exe path with language prefix */
-
-    GetModuleFileNameA(0, dll.path_tmp, MAX_PATH);
-    if ((p = strrchr( dll.path_tmp, '\\' ))) p[1] = 0;
-
-    sprintf(dll.path_manifest_exe, "%s%s", dll.path_tmp, "exe.manifest");
-    create_manifest_file(dll.path_manifest_exe, two_dll_manifest_exe_fr, -1, NULL, NULL);
-
-    strcat(dll.path_tmp, "sxs_dll\\");
-    CreateDirectoryA(dll.path_tmp, NULL);
-    sprintf(dll.path_lang,"%sfr-FR\\", dll.path_tmp);
-    CreateDirectoryA(dll.path_lang, NULL);
-    sprintf(dll.path_dll, "%s%s", dll.path_lang, "sxs_dll.dll");
-    extract_resource("dummy.dll", "TESTDLL", dll.path_dll);
-
-    sprintf(dll.path_manifest_dll, "%s%s", dll.path_lang, "sxs_dll.manifest");
-    create_manifest_file(dll.path_manifest_dll, two_dll_manifest_dll_fr, -1, NULL, NULL);
-
-    dll.context.cbSize = sizeof(ACTCTXA);
-    dll.context.lpSource = dll.path_manifest_exe;
-    dll.context.lpAssemblyDirectory = dll.path_tmp;
-    dll.context.dwFlags = ACTCTX_FLAG_ASSEMBLY_DIRECTORY_VALID;
-
-    dll.handle_context = CreateActCtxA(&dll.context);
-    ok( dll.handle_context != INVALID_HANDLE_VALUE, "CreateActCtxA failed: %ld\n", GetLastError() );
-
-    clean_sxs_info( &dll );
 }
 
 struct manifest_res_spec
@@ -4521,9 +4288,6 @@
     case 5:
        test_builtin_sxs();
        break;
-    case 6:
-       test_manifest_lang();
-       break;
     }
 }
 
@@ -4548,7 +4312,10 @@
     ret = CreateProcessA(exe, cmdline, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
     ok(ret, "Could not create process: %lu\n", GetLastError());
 
-    wait_child_process( &pi );
+    wait_child_process( pi.hProcess );
+
+    CloseHandle(pi.hThread);
+    CloseHandle(pi.hProcess);
 }
 
 static const detailed_info_t detailed_info3 =
@@ -4626,145 +4393,6 @@
     err = GetLastError();
     ok(handle == INVALID_HANDLE_VALUE, "CreateActCtxW shall fail\n");
     ok(err == ERROR_INVALID_PARAMETER, "got %lu\n", err);
-}
-
-static void test_RtlQueryInformationActiveActivationContext(void)
-{
-    ACTIVATION_CONTEXT_BASIC_INFORMATION basic;
-    ULONG_PTR cookie;
-    NTSTATUS status;
-    HANDLE context;
-    SIZE_T size;
-    BOOL ret;
-
-    if (!create_manifest_file( "test1.manifest", manifest1, -1, NULL, NULL ))
-    {
-        skip("Could not create manifest file.\n");
-        return;
-    }
-    context = test_create( "test1.manifest" );
-    ok( context != INVALID_HANDLE_VALUE, "Failed to create context, error %lu.\n", GetLastError() );
-    DeleteFileA( "test1.manifest" );
-    ret = ActivateActCtx( context, &cookie );
-    ok( ret, "ActivateActCtx failed, error %lu.\n", GetLastError() );
-
-    status = pRtlQueryInformationActiveActivationContext( ActivationContextBasicInformation, &basic,
-                                                          sizeof(basic), &size );
-    ok( status == STATUS_SUCCESS, "Got unexpected status %#lx.\n", status );
-    ok( size == sizeof(ACTIVATION_CONTEXT_BASIC_INFORMATION), "Size mismatch.\n" );
-    ok( basic.dwFlags == 0, "Got unexpected flags %#lx.\n", basic.dwFlags );
-    ok( basic.hActCtx == context, "Got unexpected handle.\n" );
-    ReleaseActCtx( basic.hActCtx );
-
-    ret = DeactivateActCtx( 0, cookie );
-    ok( ret, "DeactivateActCtx failed, error %lu.\n", GetLastError() );
-    ReleaseActCtx( context );
-}
-
-static void test_RtlActivateActivationContextUnsafeFast(void)
-{
-    RTL_CALLER_ALLOCATED_ACTIVATION_CONTEXT_STACK_FRAME_EXTENDED frame_extended1 = {0}, frame_extended2 = {0};
-    HANDLE context1, context2, current_context;
-    BOOL ret;
-
-    ret = GetCurrentActCtx(&current_context);
-    ok(ret, "GetCurrentActCtx failed.\n");
-    ok(!current_context, "Got unexpected handle.\n");
-
-    if (!create_manifest_file("test1.manifest", manifest1, -1, NULL, NULL))
-    {
-        skip("Could not create manifest file 1.\n");
-        return;
-    }
-    if (!create_manifest_file("test2.manifest", manifest1_1, -1, NULL, NULL))
-    {
-        skip("Could not create manifest file 2.\n");
-        DeleteFileA("test1.manifest");
-        return;
-    }
-    context1 = test_create("test1.manifest");
-    ok(context1 != INVALID_HANDLE_VALUE, "Failed to create context, error %lu.\n", GetLastError());
-    DeleteFileA("test1.manifest");
-    context2 = test_create("test2.manifest");
-    ok(context2 != INVALID_HANDLE_VALUE, "Failed to create context, error %lu.\n", GetLastError());
-    DeleteFileA("test2.manifest");
-
-    /* Size and Format don't seem to be checked */
-    frame_extended1.Size = 0;
-    frame_extended1.Format = 0xdeadbeef;
-    call_fastcall_func2( pRtlActivateActivationContextUnsafeFast, &frame_extended1, context1);
-    ok(frame_extended1.Size == 0, "Got unexpected Size %#Ix.\n", frame_extended1.Size);
-    ok(frame_extended1.Format == 0xdeadbeef, "Got unexpected Format %#lx.\n", frame_extended1.Format);
-    ok(!frame_extended1.Frame.Previous, "Got unexpected Previous.\n");
-    ok(frame_extended1.Frame.ActivationContext == context1, "Got unexpected ActivationContext.\n");
-    ok(!(frame_extended1.Frame.Flags & NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED),
-       "Got unexpected Flags %#lx.\n", frame_extended1.Frame.Flags);
-
-    ret = GetCurrentActCtx(&current_context);
-    ok(current_context == context1, "Got unexpected handle.\n");
-    ReleaseActCtx(current_context);
-
-    call_fastcall_func1( pRtlDeactivateActivationContextUnsafeFast, &frame_extended1);
-    ret = GetCurrentActCtx(&current_context);
-    ok(ret, "GetCurrentActCtx failed.\n");
-    ok(!current_context, "Got unexpected handle.\n");
-
-    /* Normal call */
-    frame_extended1.Size = sizeof(frame_extended1);
-    frame_extended1.Format = 0;
-    call_fastcall_func2( pRtlActivateActivationContextUnsafeFast, &frame_extended1, context1);
-    ok(frame_extended1.Size == sizeof(frame_extended1), "Got unexpected Size %#Ix.\n", frame_extended1.Size);
-    ok(frame_extended1.Format == 0, "Got unexpected Format %#lx.\n", frame_extended1.Format);
-    ok(!frame_extended1.Frame.Previous, "Got unexpected Previous.\n");
-    ok(frame_extended1.Frame.ActivationContext == context1, "Got unexpected ActivationContext.\n");
-    ok(!(frame_extended1.Frame.Flags & NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED),
-       "Got unexpected Flags %#lx.\n", frame_extended1.Frame.Flags);
-
-    ret = GetCurrentActCtx(&current_context);
-    ok(current_context == context1, "Got unexpected handle.\n");
-    ReleaseActCtx(current_context);
-
-    /* Activate another activation context */
-    frame_extended2.Size = sizeof(frame_extended2);
-    frame_extended2.Format = 0;
-    call_fastcall_func2( pRtlActivateActivationContextUnsafeFast, &frame_extended2, context2);
-    ok(frame_extended2.Size == sizeof(frame_extended2), "Got unexpected Size %#Ix.\n", frame_extended2.Size);
-    ok(frame_extended2.Format == 0, "Got unexpected Format %#lx.\n", frame_extended2.Format);
-    ok(frame_extended2.Frame.Previous == &frame_extended1.Frame, "Got unexpected Previous.\n");
-    ok(frame_extended2.Frame.ActivationContext == context2, "Got unexpected ActivationContext.\n");
-    ok(!(frame_extended2.Frame.Flags & NTDLL_ACTCTX_STACK_FRAME_HEAP_ALLOCATED),
-       "Got unexpected Flags %#lx.\n", frame_extended2.Frame.Flags);
-
-    ret = GetCurrentActCtx(&current_context);
-    ok(current_context == context2, "Got unexpected handle.\n");
-    ReleaseActCtx(current_context);
-
-    call_fastcall_func1( pRtlDeactivateActivationContextUnsafeFast, &frame_extended2);
-    ret = GetCurrentActCtx(&current_context);
-    ok(current_context == context1, "Got unexpected handle.\n");
-    ReleaseActCtx(current_context);
-
-    call_fastcall_func1( pRtlDeactivateActivationContextUnsafeFast, &frame_extended1);
-    ret = GetCurrentActCtx(&current_context);
-    ok(ret, "GetCurrentActCtx failed.\n");
-    ok(!current_context, "Got unexpected handle.\n");
-
-    /* Test freeing an unsafe frame. Not crashing means RtlFreeThreadActivationContextStack() must
-     * be able to handle it, most likely by checking the frame flags */
-    frame_extended1.Size = sizeof(frame_extended1);
-    call_fastcall_func2( pRtlActivateActivationContextUnsafeFast, &frame_extended1, context1);
-
-    ret = GetCurrentActCtx(&current_context);
-    ok(current_context == context1, "Got unexpected handle.\n");
-    ReleaseActCtx(current_context);
-
-    pRtlFreeThreadActivationContextStack();
-    ret = GetCurrentActCtx(&current_context);
-    ok(ret, "GetCurrentActCtx failed.\n");
-    ok(!current_context, "Got unexpected handle.\n");
-
-    ReleaseActCtx(context2);
-    ReleaseActCtx(context1);
 }
 
 START_TEST(actctx)
@@ -4806,7 +4434,9 @@
     run_child_process();
     test_compatibility();
     test_settings();
-    test_RtlQueryInformationActiveActivationContext();
-    test_RtlActivateActivationContextUnsafeFast();
-    for (int i = 1; i <= 6; i++) run_child_process_two_dll(i);
-}
+    run_child_process_two_dll(1);
+    run_child_process_two_dll(2);
+    run_child_process_two_dll(3);
+    run_child_process_two_dll(4);
+    run_child_process_two_dll(5);
+}
