diff --git a/dlls/mshtml/tests/es5.js b/dlls/mshtml/tests/es5.js
--- a/dlls/mshtml/tests/es5.js
+++ b/dlls/mshtml/tests/es5.js
@@ -870,6 +870,16 @@
     test_accessor_prop_desc(child.prototype, "funcprop_prot", desc);
     ok(obj.funcprop_prot(100) === 10, "obj.funcprop_prot() = " + obj.funcprop_prot(100));
 
+    (function() {
+        ok(arguments.length === 3, "arguments.length = " + arguments.length);
+        ok(arguments[0] === 1, "arguments[0] = " + arguments[0]);
+        ok(arguments[1] === 2, "arguments[1] = " + arguments[1]);
+        ok(arguments[2] === 3, "arguments[2] = " + arguments[2]);
+        Object.defineProperty(arguments, "1", {value: "foobar", writable: false, enumerable: true, configurable: false});
+        test_own_data_prop_desc(arguments, "1", false, true, false);
+        ok(arguments[1] === "foobar", "arguments[1] after defineProperty = " + arguments[1]);
+    })(1, 2, 3);
+
     expect_exception(function() {
         Object.defineProperty(null, "funcprop", desc);
     }, JS_E_OBJECT_EXPECTED);
@@ -2109,98 +2119,114 @@
     ok(r === 11.875, "buf.slice(-9) view(1).getFloat64(0) returned " + r);
 
     for(i = 0; i < types.length; i++) {
-        var name = types[i][0] + "Array", constr = window[name], typeSz = types[i][1];
-
-        ok(constr.length === 1, name + ".length = " + constr.length);
-        ok(constr.prototype.set.length === 2, name + ".prototype.set.length = " + constr.prototype.set.length);
-        ok(constr.prototype.subarray.length === 2, name + ".prototype.subarray.length = " + constr.prototype.subarray.length);
-
-        test_own_data_prop_desc(constr, "BYTES_PER_ELEMENT", false, false, false);
-        ok(constr.BYTES_PER_ELEMENT === typeSz, name + ".BYTES_PER_ELEMENT = " + constr.BYTES_PER_ELEMENT);
-
-        r = Object.getPrototypeOf(constr);
-        ok(r === Function.prototype, name + "'s prototype is not Function.prototype: " + r);
-        test_own_data_prop_desc(constr.prototype, "byteLength", false, false, false);
-        test_own_data_prop_desc(constr.prototype, "byteOffset", false, false, false);
-        test_own_data_prop_desc(constr.prototype, "length", false, false, false);
-        test_own_data_prop_desc(constr.prototype, "buffer", false, false, false);
+        var arrType = types[i][0] + "Array", typeSz = types[i][1];
+        test_own_props(arrType, [ "BYTES_PER_ELEMENT" ]);
+        test_not_own_props(arrType, [ "from", "of" ]);
+        test_own_props(arrType + ".prototype", [ "buffer", "byteLength", "byteOffset", "length", "set", "subarray" ]);
+        test_not_own_props(arrType + ".prototype", [
+            "at", "copyWithin", "entries", "every", "fill", "filter", "find", "findIndex", "forEach",
+            "includes", "indexOf", "join", "keys", "lastIndexOf", "map", "reduce", "reduceRight",
+            "reverse", "slice", "some", "sort", "toLocaleString", "toString", "values"
+        ]);
+
+        arr = eval(arrType);
+        test_own_data_prop_desc(arr, "BYTES_PER_ELEMENT", false, false, false);
+        ok(arr.BYTES_PER_ELEMENT === typeSz, arrType + ".BYTES_PER_ELEMENT = " + arr.BYTES_PER_ELEMENT);
+        r = arr.length;
+        ok(r === 1, arrType + ".length = " + r);
+        r = arr.prototype.set.length;
+        ok(r === 2, arrType + ".prototype.set.length = " + r);
+        r = arr.prototype.subarray.length;
+        ok(r === 2, arrType + ".prototype.subarray.length = " + r);
+
+        r = eval("Object.getPrototypeOf(" + arrType + ")");
+        ok(r === Function.prototype, arrType + "'s prototype is not Function.prototype: " + r);
+        r = eval("Object.getPrototypeOf(" + arrType + ".prototype)");
+        ok(r === Object.prototype, arrType + ".prototype's prototype is not Object.prototype: " + r);
+        r = eval("Object.prototype.toString.call(new " + arrType + "(3))");
+        ok(r === "[object " + arrType + "]", "Object toString(new " + arrType + "(3)) = " + r);
+        r = eval(arrType + ".prototype");
+        test_own_data_prop_desc(r, "byteLength", false, false, false);
+        test_own_data_prop_desc(r, "byteOffset", false, false, false);
+        test_own_data_prop_desc(r, "length", false, false, false);
+        test_own_data_prop_desc(r, "buffer", false, false, false);
 
         buf = ArrayBuffer(34);
         try {
-            new constr(-1);
-            ok(false, "new " + name + "(-1) did not throw exception");
+            eval("new " + arrType + "(-1)");
+            ok(false, "new " + arrType + "(-1) did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + name + "(-1) threw " + n);
-        }
-        try {
-            new constr(buf, -1);
-            ok(false, "new " + name + "(buf, -1) did not throw exception");
+            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + arrType + "(-1) threw " + n);
+        }
+        try {
+            eval("new " + arrType + "(buf, -1)");
+            ok(false, "new " + arrType + "(buf, -1) did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + name + "(buf, -1) threw " + n);
-        }
-        try {
-            new constr(buf, 36);
-            ok(false, "new " + name + "(buf, 36) did not throw exception");
+            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + arrType + "(buf, -1) threw " + n);
+        }
+        try {
+            eval("new " + arrType + "(buf, 36)");
+            ok(false, "new " + arrType + "(buf, 36) did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + name + "(buf, 36) threw " + n);
-        }
-        try {
-            new constr(buf, 32, 4);
-            ok(false, "new " + name + "(buf, 32, 4) did not throw exception");
+            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + arrType + "(buf, 36) threw " + n);
+        }
+        try {
+            eval("new " + arrType + "(buf, 32, 4)");
+            ok(false, "new " + arrType + "(buf, 32, 4) did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + name + "(buf, 32, 4) threw " + n);
-        }
-        try {
-            new constr('9');
-            ok(false, "new " + name + "('9') did not throw exception");
+            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + arrType + "(buf, 32, 4) threw " + n);
+        }
+        try {
+            eval("new " + arrType + "('9')");
+            ok(false, "new " + arrType + "('9') did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_BAD_CTOR_ARG, "new " + name + "('9') threw " + n);
-        }
-        try {
-            new constr(null);
-            ok(false, "new " + name + "(null) did not throw exception");
+            ok(n === JS_E_TYPEDARRAY_BAD_CTOR_ARG, "new " + arrType + "('9') threw " + n);
+        }
+        try {
+            eval("new " + arrType + "(null)");
+            ok(false, "new " + arrType + "(null) did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_BAD_CTOR_ARG, "new " + name + "(null) threw " + n);
-        }
-        try {
-            new constr({});
-            ok(false, "new " + name + "({}) did not throw exception");
+            ok(n === JS_E_TYPEDARRAY_BAD_CTOR_ARG, "new " + arrType + "(null) threw " + n);
+        }
+        try {
+            eval("new " + arrType + "({})");
+            ok(false, "new " + arrType + "({}) did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_BAD_CTOR_ARG, "new " + name + "({}) threw " + n);
+            ok(n === JS_E_TYPEDARRAY_BAD_CTOR_ARG, "new " + arrType + "({}) threw " + n);
         }
         if(typeSz > 1) {
             /* test misalignment */
             var a = typeSz >>> 1;
             try {
-                new constr(buf, a, 1);
-                ok(false, "new " + name + "(buf, " + a + ", 1) did not throw exception");
+                eval("new " + arrType + "(buf, a, 1)");
+                ok(false, "new " + arrType + "(buf, " + a + ", 1) did not throw exception");
             }catch(ex) {
                 var n = ex.number >>> 0;
-                ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + name + "(buf, " + a + ", 1) threw " + n);
+                ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + arrType + "(buf, " + a + ", 1) threw " + n);
             }
             a += typeSz;
             var b = new ArrayBuffer(a);
             try {
-                new constr(b);
-                ok(false, "new " + name + "(new ArrayBuffer(" + a + ")) did not throw exception");
+                eval("new " + arrType + "(b)");
+                ok(false, "new " + arrType + "(new ArrayBuffer(" + a + ")) did not throw exception");
             }catch(ex) {
                 var n = ex.number >>> 0;
-                ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + name + "(new ArrayBuffer(" + a + ")) threw " + n);
+                ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, "new " + arrType + "(new ArrayBuffer(" + a + ")) threw " + n);
             }
         }
 
-        arr = new constr();
-        ok(arr.byteLength === 0, name + "().byteLength = " + arr.byteLength);
-        ok(arr.byteOffset === 0, name + "().byteOffset = " + arr.byteOffset);
-        ok(arr.length === 0, name + "().length = " + arr.length);
-        ok(arr.buffer.byteLength === 0, name + "().buffer.byteLength = " + arr.buffer.byteLength);
+        arr = eval("new " + arrType + "()");
+        ok(arr.byteLength === 0, arrType + "().byteLength = " + arr.byteLength);
+        ok(arr.byteOffset === 0, arrType + "().byteOffset = " + arr.byteOffset);
+        ok(arr.length === 0, arrType + "().length = " + arr.length);
+        ok(arr.buffer.byteLength === 0, arrType + "().buffer.byteLength = " + arr.buffer.byteLength);
         test_readonly(arr, "byteLength", 0);
         test_readonly(arr, "byteOffset", 0);
         test_readonly(arr, "length", 0);
@@ -2209,95 +2235,86 @@
         test_own_data_prop_desc(arr, "length", false, false, false);
         test_own_data_prop_desc(arr, "buffer", false, false, false);
 
-        ok(ArrayBuffer.isView(arr) === true, "ArrayBuffer.isView(" + name + "()) returned false");
+        ok(ArrayBuffer.isView(arr) === true, "ArrayBuffer.isView(" + arrType + "()) returned false");
         Object.freeze(arr);
-        ok(Object.isFrozen(arr) === true, name + "() not frozen");
-
-        arr = constr(9.1);
-        ok(arr.byteLength === 9 * typeSz, name + "(9.1).byteLength = " + arr.byteLength);
-        ok(arr.byteOffset === 0, name + "(9.1).byteOffset = " + arr.byteOffset);
-        ok(arr.length === 9, name + "(9.1).length = " + arr.length);
-        ok(arr.buffer.byteLength === arr.byteLength, name + "(9.1).buffer.byteLength = " + arr.buffer.byteLength);
+        ok(Object.isFrozen(arr) === true, arrType + "() not frozen");
+
+        arr = eval(arrType + "(9.1)");
+        ok(arr.byteLength === 9 * typeSz, arrType + "(9.1).byteLength = " + arr.byteLength);
+        ok(arr.byteOffset === 0, arrType + "(9.1).byteOffset = " + arr.byteOffset);
+        ok(arr.length === 9, arrType + "(9.1).length = " + arr.length);
+        ok(arr.buffer.byteLength === arr.byteLength, arrType + "(9.1).buffer.byteLength = " + arr.buffer.byteLength);
         for(var j = 0; j < 9; j++)
             ok(arr[j] === 0, "arr[" + j + "] = " + arr[j]);
         arr[5] = 42;
-        ok(arr[5] === 42, name + "(9.1)[5] = " + arr[5]);
+        ok(arr[5] === 42, arrType + "(9.1)[5] = " + arr[5]);
         arr[9] = 50;
-        ok(arr[9] === undefined, name + "(9.1)[9] = " + arr[9]);
-
-        constr.prototype[6] = "foo";
-        r = constr.prototype[6];
-        ok(r === undefined, name + ".prototype[6] = " + r);
-        ok(arr[6] === 0, name + "(9.1)[6] after set in prototype = " + arr[6]);
+        ok(arr[9] === undefined, arrType + "(9.1)[9] = " + arr[9]);
+
+        eval(arrType + ".prototype[6] = 'foo'");
+        r = eval(arrType + ".prototype[6]");
+        ok(r === undefined, arrType + ".prototype[6] = " + r);
+        ok(arr[6] === 0, arrType + "(9.1)[6] after set in prototype = " + arr[6]);
         arr[6] = 0;
-        ok(arr.hasOwnProperty("6"), "'6' not a property of " + name + "(9.1)[6]");
+        ok(Object.prototype.hasOwnProperty.call(arr, "6"), "'6' not a property of " + arrType + "(9.1)[6]");
         test_own_data_prop_desc(arr, "6", true, true, false);
         r = (delete arr[6]);
-        ok(r === false, "delete " + name + "(9.1)[6] returned " + r);
+        ok(r === false, "delete " + arrType + "(9.1)[6] returned " + r);
         try {
             Object.defineProperty(arr, "6", {writable: false, enumerable: false, configurable: true, value: 10});
-            ok(false, "redefining " + name + "(9.1)[6] with different flags did not throw exception");
+            ok(false, "redefining " + arrType + "(9.1)[6] with different flags did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_NONCONFIGURABLE_REDEFINED, "redefining " + name + "(9.1)[6] with different flags threw " + n);
+            ok(n === JS_E_NONCONFIGURABLE_REDEFINED, "redefining " + arrType + "(9.1)[6] with different flags threw " + n);
         }
         Object.defineProperty(arr, "6", {writable: true, enumerable: true, configurable: false, value: 10});
-        ok(arr[6] === 10, name + "(9.1)[6] after definition = " + arr[6]);
+        ok(arr[6] === 10, arrType + "(9.1)[6] after definition = " + arr[6]);
         Object.defineProperty(arr, "6", {writable: true, enumerable: true, configurable: false, value: "foo"});
-        if(name.substr(0, 5) === "Float")
-            todo_wine.
-            ok(arr[6] !== arr[6] /* NaN */, name + "(9.1)[6] after definition to string = " + arr[6]);
+        if(arrType.substr(0, 5) === "Float")
+            ok(arr[6] !== arr[6] /* NaN */, arrType + "(9.1)[6] after definition to string = " + arr[6]);
         else
-            todo_wine.
-            ok(arr[6] === 0, name + "(9.1)[6] after definition to string = " + arr[6]);
-
-        constr.prototype[100] = "foobar";
-        r = constr.prototype[100];
-        ok(r === undefined, name + ".prototype[100] = " + r);
-        ok(arr[100] === undefined, name + "(9.1)[100] after set in prototype = " + arr[100]);
+            ok(arr[6] === 0, arrType + "(9.1)[6] after definition to string = " + arr[6]);
+
+        eval(arrType + ".prototype[100] = 'foobar'");
+        r = eval(arrType + ".prototype[100]");
+        ok(r === undefined, arrType + ".prototype[100] = " + r);
+        ok(arr[100] === undefined, arrType + "(9.1)[100] after set in prototype = " + arr[100]);
         arr[100] = 0;
-        ok(arr[100] === undefined, name + "(9.1)[100] after set to zero = " + arr[100]);
-        todo_wine.
-        ok(!arr.hasOwnProperty("100"), "'100' is a property of " + name + "(9.1)[100]");
+        ok(arr[100] === undefined, arrType + "(9.1)[100] after set to zero = " + arr[100]);
+        ok(!Object.prototype.hasOwnProperty.call(arr, "100"), "'100' is a property of " + arrType + "(9.1)[100]");
         r = (delete arr[100]);
-        ok(r === false, "delete " + name + "(9.1)[100] returned " + r);
+        ok(r === false, "delete " + arrType + "(9.1)[100] returned " + r);
         try {
             Object.defineProperty(arr, "100", {writable: false, enumerable: false, configurable: true, value: 10});
-            ok(false, "redefining " + name + "(9.1)[100] with different flags did not throw exception");
+            ok(false, "redefining " + arrType + "(9.1)[100] with different flags did not throw exception");
         }catch(ex) {
             var n = ex.number >>> 0;
-            ok(n === JS_E_NONCONFIGURABLE_REDEFINED, "redefining " + name + "(9.1)[100] with different flags threw " + n);
+            ok(n === JS_E_NONCONFIGURABLE_REDEFINED, "redefining " + arrType + "(9.1)[100] with different flags threw " + n);
         }
         Object.defineProperty(arr, "100", {writable: true, enumerable: true, configurable: false, value: 10});
-        todo_wine.
-        ok(arr[100] === undefined, name + "(9.1)[100] after defined to 10 = " + arr[100]);
-        todo_wine.
-        ok(!arr.hasOwnProperty("100"), "'100' is a property of " + name + "(9.1)[100] after definition");
-        todo_wine.
-        ok(arr[100] === undefined, name + "(9.1)[100] after definition = " + arr[100]);
+        ok(arr[100] === undefined, arrType + "(9.1)[100] after defined to 10 = " + arr[100]);
+        ok(!Object.prototype.hasOwnProperty.call(arr, "100"), "'100' is a property of " + arrType + "(9.1)[100] after definition");
+        ok(arr[100] === undefined, arrType + "(9.1)[100] after definition = " + arr[100]);
 
         r = 0;
         for(var idx in arr) {
-            todo_wine_if(r == 10).
-            ok(idx === ""+r, name + "(9.1) enum idx " + r + " = " + idx);
+            ok(idx === ""+r, arrType + "(9.1) enum idx " + r + " = " + idx);
             r++;
         }
-        todo_wine.
-        ok(r === 9, name + "(9.1) enum did " + r + " iterations");
-
-        constr.prototype[-1] = "barfoo";
-        r = constr.prototype[-1];
-        ok(r === "barfoo", name + ".prototype[-1] = " + r);
-        ok(arr[-1] === "barfoo", name + "(9.1)[-1] after set in prototype = " + arr[-1]);
-
-        constr.prototype.foo = "bar";
-        r = constr.prototype.foo;
-        ok(r === "bar", name + ".prototype.foo = " + r);
-        ok(arr.foo === "bar", name + "(9.1).foo after set in prototype = " + arr.foo);
+        ok(r === 9, arrType + "(9.1) enum did " + r + " iterations");
+
+        eval(arrType + ".prototype[-1] = 'barfoo'");
+        r = eval(arrType + ".prototype[-1]");
+        ok(r === "barfoo", arrType + ".prototype[-1] = " + r);
+        ok(arr[-1] === "barfoo", arrType + "(9.1)[-1] after set in prototype = " + arr[-1]);
+
+        eval(arrType + ".prototype.foo = 'bar'");
+        r = eval(arrType + ".prototype.foo = 'bar'");
+        ok(r === "bar", arrType + ".prototype.foo = " + r);
+        ok(arr.foo === "bar", arrType + "(9.1).foo after set in prototype = " + arr.foo);
         Object.freeze(arr);
-        todo_wine.
-        ok(Object.isFrozen(arr) === true, name + "(9.1) not frozen");
-        arr = constr.prototype;
+        ok(Object.isFrozen(arr) === true, arrType + "(9.1) not frozen");
+        arr = eval(arrType + ".prototype");
         delete arr[-1];
         delete arr.foo;
 
@@ -2305,151 +2322,21 @@
         arr2[0] = 1.5;
         arr2[1] = '3';
         arr2[3] = 12;
-        arr = constr(arr2);
-        r = name + "(array-like object)";
-        ok(arr.byteLength === 4 * typeSz, r + ".byteLength = " + arr.byteLength);
-        ok(arr.byteOffset === 0, r + ".byteOffset = " + arr.byteOffset);
-        ok(arr.length === 4, r + ".length = " + arr.length);
+        var name = arrType + "(array-like object)";
+        arr = eval(arrType + "(arr2)");
+        ok(arr.byteLength === 4 * typeSz, name + ".byteLength = " + arr.byteLength);
+        ok(arr.byteOffset === 0, name + ".byteOffset = " + arr.byteOffset);
+        ok(arr.length === 4, name + ".length = " + arr.length);
         if(isNaN(arr[2])) {
-            ok(arr[0] === 1.5, r + "[0] = " + arr[0]);
-            ok(arr[1] === 3,   r + "[1] = " + arr[1]);
-            ok(arr[3] === 12,  r + "[3] = " + arr[3]);
+            ok(arr[0] === 1.5, name + "[0] = " + arr[0]);
+            ok(arr[1] === 3,   name + "[1] = " + arr[1]);
+            ok(arr[3] === 12,  name + "[3] = " + arr[3]);
         }else
             for(var j = 0; j < 4; j++)
-                ok(arr[j] === [1, 3, 0, 12][j], r + "[" + j + "] = " + arr[j]);
-
-        arr = constr(10);
-        r = name + "(10)";
-        try {
-            arr.subarray.call(null, 0);
-            ok(false, r + ": calling subarray with null context did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_NOT_TYPEDARRAY, r + ": calling subarray with null context threw " + n);
-        }
-        try {
-            arr.subarray.call({}, 0);
-            ok(false, r + ": calling subarray with an object context did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_NOT_TYPEDARRAY, r + ": calling subarray with an object context threw " + n);
-        }
-        try {
-            arr.subarray();
-            ok(false, r + " subarray() did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_SUBARRAY, r + " subarray() threw " + n);
-        }
-        arr2 = arr.subarray(4);
-        ok(arr2.byteLength === 6 * typeSz, r + ".subarray(4).byteLength = " + arr2.byteLength);
-        ok(arr2.byteOffset === 4 * typeSz, r + ".subarray(4).byteOffset = " + arr2.byteOffset);
-        ok(arr2.length === 6, r + ".subarray(4).length = " + arr2.length);
-        ok(arr2.buffer === arr.buffer, r + ".subarray(4).buffer = " + arr2.buffer);
-        arr2 = arr.subarray(4, 2);
-        ok(arr2.byteLength === 0, r + ".subarray(4, 2).byteLength = " + arr2.byteLength);
-        ok(arr2.byteOffset === 4 * typeSz, r + ".subarray(4, 2).byteOffset = " + arr2.byteOffset);
-        ok(arr2.length === 0, r + ".subarray(4, 2).length = " + arr2.length);
-        ok(arr2.buffer === arr.buffer, r + ".subarray(4, 2).buffer = " + arr2.buffer);
-        arr2 = arr.subarray(-3, 100);
-        ok(arr2.byteLength === 3 * typeSz, r + ".subarray(-3, 100).byteLength = " + arr2.byteLength);
-        ok(arr2.byteOffset === 7 * typeSz, r + ".subarray(-3, 100).byteOffset = " + arr2.byteOffset);
-        ok(arr2.length === 3, r + ".subarray(-3, 100).length = " + arr2.length);
-        ok(arr2.buffer === arr.buffer, r + ".subarray(-3, 100).buffer = " + arr2.buffer);
-        arr2 = arr.subarray(42, -1);
-        ok(arr2.byteLength === 0, r + ".subarray(42, -1).byteLength = " + arr2.byteLength);
-        ok(arr2.byteOffset === 10 * typeSz, r + ".subarray(42, -1).byteOffset = " + arr2.byteOffset);
-        ok(arr2.length === 0, r + ".subarray(42, -1).length = " + arr2.length);
-        ok(arr2.buffer === arr.buffer, r + ".subarray(42, -1).buffer = " + arr2.buffer);
-        arr2 = arr.subarray(2, -3);
-        ok(arr2.byteLength === 5 * typeSz, r + ".subarray(2, -3).byteLength = " + arr2.byteLength);
-        ok(arr2.byteOffset === 2 * typeSz, r + ".subarray(2, -3).byteOffset = " + arr2.byteOffset);
-        ok(arr2.length === 5, r + ".subarray(2, -3).length = " + arr2.length);
-        ok(arr2.buffer === arr.buffer, r + ".subarray(2, -3).buffer = " + arr2.buffer);
-
-        try {
-            arr.set.call(null, [1]);
-            ok(false, name + ": calling set with null context did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_NOT_TYPEDARRAY, name + ": calling set with null context threw " + n);
-        }
-        try {
-            arr.set.call({}, [1]);
-            ok(false, name + ": calling set with an object context did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_NOT_TYPEDARRAY, name + ": calling set with an object context threw " + n);
-        }
-        try {
-            arr.set();
-            ok(false, name + ".set() did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_SOURCE, name + ".set() threw " + n);
-        }
-        try {
-            arr.set(null);
-            ok(false, name + ".set(null) did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_SOURCE, name + ".set(null) threw " + n);
-        }
-        try {
-            arr.set([1,2,3], 8);
-            ok(false, name + ".set([1,2,3], 8) did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, name + ".set([1,2,3], 8) threw " + n);
-        }
-        try {
-            arr.set([99], -3);
-            ok(false, name + ".set([99], -3) did not throw exception");
-        }catch(ex) {
-            var n = ex.number >>> 0;
-            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, name + ".set([99], -3) threw " + n);
-        }
-
-        r = arr.set(5);
-        ok(r === undefined, name + ".set(5) returned " + r);
-        for(var j = 0; j < 10; j++)
-            ok(arr[j] === 0, name + ".set(5): arr[" + j + "] = " + arr[j]);
-
-        r = arr.set({});
-        ok(r === undefined, name + ".set({}) returned " + r);
-        for(var j = 0; j < 10; j++)
-            ok(arr[j] === 0, name + ".set({}): arr[" + j + "] = " + arr[j]);
-
-        r = arr.set("12");
-        ok(r === undefined, name + ".set('12') returned " + r);
-        for(var j = 0; j < 10; j++)
-            ok(arr[j] === [ 1, 2, 0, 0, 0, 0, 0, 0, 0, 0 ][j], name + ".set('12'): arr[" + j + "] = " + arr[j]);
-
-        arr2 = { length: 2 };
-        arr2[0] = 9;
-        arr2[1] = 7;
-        r = arr.set(arr2);
-        ok(r === undefined, name + ".set(array-like obj) returned " + r);
-        for(var j = 0; j < 10; j++)
-            ok(arr[j] === [ 9, 7, 0, 0, 0, 0, 0, 0, 0, 0 ][j], name + ".set(array-like obj): arr[" + j + "] = " + arr[j]);
-
-        r = arr.set([12, 10, 11], 3);
-        ok(r === undefined, name + ".set([12, 10, 11], 3) returned " + r);
-        for(var j = 0; j < 10; j++)
-            ok(arr[j] === [ 9, 7, 0, 12, 10, 11, 0, 0, 0, 0 ][j], name + ".set([12, 10, 11], 3): arr[" + j + "] = " + arr[j]);
-
-        r = arr.set(arr.subarray(4, 6), 5);
-        ok(r === undefined, name + ".set(arr.subarray(4, 2), 5) returned " + r);
-        for(var j = 0; j < 10; j++)
-            ok(arr[j] === [ 9, 7, 0, 12, 10, 10, 11, 0, 0, 0 ][j], name + ".set(arr.subarray(4, 2), 5): arr[" + j + "] = " + arr[j]);
-
-        r = arr.set(arr.subarray(3, 7), 2);
-        ok(r === undefined, name + ".set(arr.subarray(3, 7), 2) returned " + r);
-        for(var j = 0; j < 10; j++)
-            ok(arr[j] === [ 9, 7, 12, 10, 10, 11, 11, 0, 0, 0 ][j], name + ".set(arr.subarray(3, 7), 2): arr[" + j + "] = " + arr[j]);
-
-        arr = constr(buf, typeSz, 2);
-        name = name + "(buf, " + typeSz + ", 2)";
+                ok(arr[j] === [1, 3, 0, 12][j], name + "[" + j + "] = " + arr[j]);
+
+        name = arrType + "(buf, " + typeSz + ", 2)";
+        arr = eval(name);
         ok(arr.byteLength === 2 * typeSz, name + ".byteLength = " + arr.byteLength);
         ok(arr.byteOffset === typeSz, name + ".byteOffset = " + arr.byteOffset);
         ok(arr.length === 2, name + ".length = " + arr.length);
@@ -2461,10 +2348,9 @@
         r = view["get" + types[i][0]](typeSz, true);
         ok(r === 12, "DataView(buf).get" + types[i][0] + " after arr[0] set = " + r);
         Object.freeze(arr);
-        todo_wine.
         ok(Object.isFrozen(arr) === true, name + " not frozen");
 
-        arr2 = constr(arr);
+        arr2 = eval(arrType + "(arr)");
         ok(arr2.byteLength === arr.byteLength, name + " copy.byteLength = " + arr2.byteLength);
         ok(arr2.byteOffset === 0, name + " copy.byteOffset = " + arr2.byteOffset);
         ok(arr2.length === arr.length, name + " copy.length = " + arr2.length);
@@ -2474,6 +2360,136 @@
         ok(arr2.byteOffset === arr.byteOffset, name + " subarray(undefined, '1').byteOffset = " + arr2.byteOffset);
         ok(arr2.length === 1, name + " subarray(undefined, '1').length = " + arr2.length);
         ok(arr2.buffer === arr.buffer, name + " subarray(undefined, '1').buffer = " + arr2.buffer);
+
+        name = arrType + "(10)";
+        arr = eval(name);
+        try {
+            arr.subarray.call(null, 0);
+            ok(false, arrType + ": calling subarray with null context did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_NOT_TYPEDARRAY, arrType + ": calling subarray with null context threw " + n);
+        }
+        try {
+            arr.subarray.call({}, 0);
+            ok(false, arrType + ": calling subarray with an object context did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_NOT_TYPEDARRAY, arrType + ": calling subarray with an object context threw " + n);
+        }
+        try {
+            arr.subarray();
+            ok(false, name + " subarray() did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_TYPEDARRAY_INVALID_SUBARRAY, name + " subarray() threw " + n);
+        }
+        arr2 = arr.subarray(4);
+        ok(arr2.byteLength === 6 * typeSz, name + ".subarray(4).byteLength = " + arr2.byteLength);
+        ok(arr2.byteOffset === 4 * typeSz, name + ".subarray(4).byteOffset = " + arr2.byteOffset);
+        ok(arr2.length === 6, name + ".subarray(4).length = " + arr2.length);
+        ok(arr2.buffer === arr.buffer, name + ".subarray(4).buffer = " + arr2.buffer);
+        arr2 = arr.subarray(4, 2);
+        ok(arr2.byteLength === 0, name + ".subarray(4, 2).byteLength = " + arr2.byteLength);
+        ok(arr2.byteOffset === 4 * typeSz, name + ".subarray(4, 2).byteOffset = " + arr2.byteOffset);
+        ok(arr2.length === 0, name + ".subarray(4, 2).length = " + arr2.length);
+        ok(arr2.buffer === arr.buffer, name + ".subarray(4, 2).buffer = " + arr2.buffer);
+        arr2 = arr.subarray(-3, 100);
+        ok(arr2.byteLength === 3 * typeSz, name + ".subarray(-3, 100).byteLength = " + arr2.byteLength);
+        ok(arr2.byteOffset === 7 * typeSz, name + ".subarray(-3, 100).byteOffset = " + arr2.byteOffset);
+        ok(arr2.length === 3, name + ".subarray(-3, 100).length = " + arr2.length);
+        ok(arr2.buffer === arr.buffer, name + ".subarray(-3, 100).buffer = " + arr2.buffer);
+        arr2 = arr.subarray(42, -1);
+        ok(arr2.byteLength === 0, name + ".subarray(42, -1).byteLength = " + arr2.byteLength);
+        ok(arr2.byteOffset === 10 * typeSz, name + ".subarray(42, -1).byteOffset = " + arr2.byteOffset);
+        ok(arr2.length === 0, name + ".subarray(42, -1).length = " + arr2.length);
+        ok(arr2.buffer === arr.buffer, name + ".subarray(42, -1).buffer = " + arr2.buffer);
+        arr2 = arr.subarray(2, -3);
+        ok(arr2.byteLength === 5 * typeSz, name + ".subarray(2, -3).byteLength = " + arr2.byteLength);
+        ok(arr2.byteOffset === 2 * typeSz, name + ".subarray(2, -3).byteOffset = " + arr2.byteOffset);
+        ok(arr2.length === 5, name + ".subarray(2, -3).length = " + arr2.length);
+        ok(arr2.buffer === arr.buffer, name + ".subarray(2, -3).buffer = " + arr2.buffer);
+
+        try {
+            arr.set.call(null, [1]);
+            ok(false, arrType + ": calling set with null context did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_NOT_TYPEDARRAY, arrType + ": calling set with null context threw " + n);
+        }
+        try {
+            arr.set.call({}, [1]);
+            ok(false, arrType + ": calling set with an object context did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_NOT_TYPEDARRAY, arrType + ": calling set with an object context threw " + n);
+        }
+        try {
+            arr.set();
+            ok(false, name + ".set() did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_TYPEDARRAY_INVALID_SOURCE, name + ".set() threw " + n);
+        }
+        try {
+            arr.set(null);
+            ok(false, name + ".set(null) did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_TYPEDARRAY_INVALID_SOURCE, name + ".set(null) threw " + n);
+        }
+        try {
+            arr.set([1,2,3], 8);
+            ok(false, name + ".set([1,2,3], 8) did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, name + ".set([1,2,3], 8) threw " + n);
+        }
+        try {
+            arr.set([99], -3);
+            ok(false, name + ".set([99], -3) did not throw exception");
+        }catch(ex) {
+            var n = ex.number >>> 0;
+            ok(n === JS_E_TYPEDARRAY_INVALID_OFFSLEN, name + ".set([99], -3) threw " + n);
+        }
+
+        r = arr.set(5);
+        ok(r === undefined, name + ".set(5) returned " + r);
+        for(var j = 0; j < 10; j++)
+            ok(arr[j] === 0, name + ".set(5): arr[" + j + "] = " + arr[j]);
+
+        r = arr.set({});
+        ok(r === undefined, name + ".set({}) returned " + r);
+        for(var j = 0; j < 10; j++)
+            ok(arr[j] === 0, name + ".set({}): arr[" + j + "] = " + arr[j]);
+
+        r = arr.set("12");
+        ok(r === undefined, name + ".set('12') returned " + r);
+        for(var j = 0; j < 10; j++)
+            ok(arr[j] === [ 1, 2, 0, 0, 0, 0, 0, 0, 0, 0 ][j], name + ".set('12'): arr[" + j + "] = " + arr[j]);
+
+        arr2 = { length: 2 };
+        arr2[0] = 9;
+        arr2[1] = 7;
+        r = arr.set(arr2);
+        ok(r === undefined, name + ".set(array-like obj) returned " + r);
+        for(var j = 0; j < 10; j++)
+            ok(arr[j] === [ 9, 7, 0, 0, 0, 0, 0, 0, 0, 0 ][j], name + ".set(array-like obj): arr[" + j + "] = " + arr[j]);
+
+        r = arr.set([12, 10, 11], 3);
+        ok(r === undefined, name + ".set([12, 10, 11], 3) returned " + r);
+        for(var j = 0; j < 10; j++)
+            ok(arr[j] === [ 9, 7, 0, 12, 10, 11, 0, 0, 0, 0 ][j], name + ".set([12, 10, 11], 3): arr[" + j + "] = " + arr[j]);
+
+        r = arr.set(arr.subarray(4, 6), 5);
+        ok(r === undefined, name + ".set(arr.subarray(4, 2), 5) returned " + r);
+        for(var j = 0; j < 10; j++)
+            ok(arr[j] === [ 9, 7, 0, 12, 10, 10, 11, 0, 0, 0 ][j], name + ".set(arr.subarray(4, 2), 5): arr[" + j + "] = " + arr[j]);
+
+        r = arr.set(arr.subarray(3, 7), 2);
+        ok(r === undefined, name + ".set(arr.subarray(3, 7), 2) returned " + r);
+        for(var j = 0; j < 10; j++)
+            ok(arr[j] === [ 9, 7, 12, 10, 10, 11, 11, 0, 0, 0 ][j], name + ".set(arr.subarray(3, 7), 2): arr[" + j + "] = " + arr[j]);
     }
 
     arr = new Float32Array(3);
@@ -2491,41 +2507,17 @@
     ok(arr[0] === 1.125, "arr[0] after arr2[0] changed = " + arr[0]);
     ok(arr2[0] === 100, "arr2[0] after change = " + arr2[0]);
 
-    arr = new Int8Array(2);
-    arr[0] = -413;
-    arr[1] = 261;
-    ok(arr[0] == 99, "8-bit arr[0] after overflow = " + arr[0]);
-    ok(arr[1] == 5, "8-bit arr[1] after overflow = " + arr[1]);
-
-    arr = new Uint8Array(2);
-    arr[0] = -2;
-    arr[1] = 258;
-    ok(arr[0] == 254, "8-bit unsigned arr[0] after overflow = " + arr[0]);
-    ok(arr[1] == 2, "8-bit unsigned arr[1] after overflow = " + arr[1]);
-
     arr = new Int16Array(2);
     arr[0] = 65535;
     arr[1] = -65535;
     ok(arr[0] == -1, "16-bit arr[0] after overflow = " + arr[0]);
     ok(arr[1] == 1, "16-bit arr[1] after overflow = " + arr[1]);
 
-    arr = new Uint16Array(2);
-    arr[0] = -65542;
-    arr[1] = 123456;
-    ok(arr[0] == 65530, "16-bit unsigned arr[0] after overflow = " + arr[0]);
-    ok(arr[1] == 57920, "16-bit unsigned arr[1] after overflow = " + arr[1]);
-
-    arr = new Int32Array(2);
-    arr[0] = -9876543210;
-    arr[1] = 4294967999;
-    ok(arr[0] == -1286608618, "32-bit arr[0] after overflow = " + arr[0]);
-    ok(arr[1] == 703, "32-bit arr[1] after overflow = " + arr[1]);
-
-    arr = new Uint32Array(2);
-    arr[0] = -4444444444;
-    arr[1] = 33333333333;
-    ok(arr[0] == 4145490148, "32-bit unsigned arr[0] after overflow = " + arr[0]);
-    ok(arr[1] == 3268562261, "32-bit unsigned arr[1] after overflow = " + arr[1]);
+    arr = new Uint8Array(2);
+    arr[0] = -2;
+    arr[1] = 258;
+    ok(arr[0] == 254, "8-bit arr[0] after overflow = " + arr[0]);
+    ok(arr[1] == 2, "8-bit arr[1] after overflow = " + arr[1]);
 
     arr = new Int8Array(12);
     arr.set([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);
@@ -2542,9 +2534,6 @@
         ok(arr2[j] === j + 2, "arr with overlap[" + j + "] = " + arr[j]);
 
     /* methods are incompatible, even though thrown error is not explicit */
-    arr = new Int8Array();
-    arr2 = new Int32Array();
-
     ok(Uint16Array.prototype.subarray !== Int32Array.prototype.subarray, "Uint16Array and Int32Array have same subarray methods");
     ok(Int8Array.prototype.set !== Float32Array.prototype.set, "Int8Array and Float32Array have same set methods");
     try {
@@ -2760,8 +2749,6 @@
         [137].forEach(external.testHostContext(false), tests[i]);
         Function.prototype.apply.call(external.testHostContext(false), tests[i], [137, 0, {}]);
     }
-
-    i = external.objWithMethod.func(42);
 });
 
 sync_test("head_setter", function() {
@@ -3088,81 +3075,45 @@
     })();
 });
 
-sync_test("input validation", function() {
-    var fired, elem = document.createElement("input");
-    elem.type = "number";
-    elem.setAttribute("min", "1");
-    elem.setAttribute("max", "4");
-    elem.addEventListener("invalid", function(e) {
-        ok(e.target === elem, "unexpected target " + e.target);
-        fired = true;
-    });
-    fired = false;
-    elem.value = 1;
-    ok(elem.checkValidity() === true, "input number (1-4) with value 1: invalid");
-    ok(fired === false, "input number (1-4) with value 1 fired invalid event");
-    fired = false;
-    elem.value = 0;
-    ok(elem.checkValidity() === false, "input number (1-4) with value 0: valid");
-    ok(fired === true, "input number (1-4) with value 0 did not fire invalid event");
-    fired = false;
-    elem.value = 5;
-    ok(elem.checkValidity() === false, "input number (1-4) with value 5: valid");
-    ok(fired === true, "input number (1-4) with value 5 did not fire invalid event");
-});
-
 sync_test("perf toJSON", function() {
-    var json, objs = [ performance, performance.navigation, performance.timing ];
-    var non_props = [ "constructor", "TYPE_BACK_FORWARD", "TYPE_NAVIGATE", "TYPE_RELOAD", "TYPE_RESERVED" ];
-
-    for(var i = 0; i < objs.length; i++) {
-        var desc, prop, proto = Object.getPrototypeOf(objs[i]), props = Object.getOwnPropertyNames(proto);
-        var name = Object.prototype.toString.call(objs[i]).slice(8, -1);
-
-        Object.defineProperty(objs[i], "foobar", {writable: true, enumerable: true, configurable: true, value: 1});
-        Object.defineProperty(proto, "barfoo", {writable: true, enumerable: true, configurable: true, value: 3});
-        json = objs[i].toJSON();
+    var tests = [
+        [ "performance", "navigation", "timing" ],
+        [ "performance.navigation", "redirectCount", "type" ],
+        [ "performance.timing", "connectEnd", "connectStart", "domComplete", "domContentLoadedEventEnd",
+          "domContentLoadedEventStart", "domInteractive", "domLoading", "domainLookupEnd", "domainLookupStart",
+          "fetchStart", "loadEventEnd", "loadEventStart", "msFirstPaint", "navigationStart", "redirectEnd",
+          "redirectStart", "requestStart", "responseEnd", "responseStart", "unloadEventEnd", "unloadEventStart" ]
+    ];
+
+    for(var i = 0; i < tests.length; i++) {
+        var desc, name = tests[i][0], obj = eval("window." + name), json;
+
+        Object.defineProperty(obj, "foobar", {writable: true, enumerable: true, configurable: true, value: 1});
+        Object.defineProperty(Object.getPrototypeOf(obj), "barfoo", {writable: true, enumerable: true, configurable: true, value: 3});
+        json = obj.toJSON();
 
         ok(Object.getPrototypeOf(json) === Object.prototype, "prototype of " + name + ".toJSON() != Object.prototype");
         ok(typeof json === "object", "typeof " + name + ".toJSON() != object");
-
-        for(var j = 0; j < non_props.length; j++) {
-            var idx = props.indexOf(non_props[j]);
-            if(idx !== -1)
-                props.splice(idx, 1);
-        }
-
-        for(var j = 0; j < props.length; j++) {
-            if(typeof(proto[props[j]]) === "function")
-                ok(!(props[j] in json), props[j] + " in " + name + ".toJSON()");
-            else {
-                test_own_data_prop_desc(json, props[j], true, true, true);
-                prop = props[j];
-            }
-        }
-
-        for(var j = 0; j < non_props.length; j++)
-            ok(!json.hasOwnProperty(non_props[j]), non_props[j] + " in " + name + ".toJSON()");
+        for(var j = 1; j < tests[i].length; j++) {
+            desc = Object.getOwnPropertyDescriptor(json, tests[i][j]);
+            ok(json.hasOwnProperty(tests[i][j]), name + ".toJSON() does not have " + tests[i][j]);
+            ok(desc.writable === true, name + ".toJSON()." + tests[i][j] + " not writable");
+            ok(desc.enumerable === true, name + ".toJSON()." + tests[i][j] + " not enumerable");
+            ok(desc.configurable === true, name + ".toJSON()." + tests[i][j] + " not configurable");
+        }
         ok(!("foobar" in json), "foobar in " + name + ".toJSON()");
         ok(!("barfoo" in json), "barfoo in " + name + ".toJSON()");
 
-        delete objs[i].foobar;
-        delete proto.barfoo;
-
-        // test delete a builtin from the prototype and toJSON after
-        desc = Object.getOwnPropertyDescriptor(proto, prop);
-        delete proto[prop];
-        ok(!(prop in objs[i]), prop + " in " + name + " after delete");
-        json = objs[i].toJSON();
-        ok(json.hasOwnProperty(prop), name + ".toJSON() does not have " + prop + " after delete");
-        Object.defineProperty(proto, prop, desc);
-    }
-
-    json = performance.toJSON();
-    ok(typeof json.navigation === "object", "JSON'd performance's navigation type = " + typeof json.navigation);
-    ok(typeof json.navigation.redirectCount === "number", "JSON'd performance's navigation.redirectCount type = " + typeof json.navigation.redirectCount);
-    ok(Object.getPrototypeOf(json.navigation) === Object.prototype, "JSON'd performance's navigation prototype != Object.prototype");
-    ok(!("toJSON" in json.navigation), "toJSON in JSON'd performance's navigation");
+        delete obj.foobar;
+        delete Object.getPrototypeOf(obj).barfoo;
+
+        desc = Object.getOwnPropertyDescriptor(Object.getPrototypeOf(obj), tests[i][1]);
+        delete Object.getPrototypeOf(obj)[tests[i][1]];
+        ok(!(tests[i][1] in obj), tests[i][1] + " in " + name + " after delete");
+        json = obj.toJSON();
+        ok(json.hasOwnProperty(tests[i][1]), name + ".toJSON() does not have " + tests[i][1] + " after delete");
+        Object.defineProperty(Object.getPrototypeOf(obj), tests[i][1], desc);
+    }
 });
 
 sync_test("console", function() {
@@ -3545,25 +3496,12 @@
         ok(!(doc.implementation instanceof Object), "doc.implementation is an instance of Object");
         ok(!(doc.implementation instanceof DOMImplementation), "doc.implementation is an instance of DOMImplementation");
         ok(doc.implementation instanceof iframe.contentWindow.DOMImplementation, "doc.implementation is not an instance of iframe's DOMImplementation");
-        ok(Object.getPrototypeOf(doc) !== Object.getPrototypeOf(document), "doc's prototype same as doc prototype");
-        ok(Object.getPrototypeOf(doc) === iframe.contentWindow.HTMLDocument.prototype, "doc's prototype not iframe's HTMLDocument.prototype");
 
         doc = doc.implementation.createHTMLDocument("test");
         ok(!(doc instanceof Object), "created iframe doc is an instance of Object");
         ok(!(doc.implementation instanceof Object), "created iframe doc.implementation is an instance of Object");
         ok(!(doc.implementation instanceof DOMImplementation), "created iframe doc.implementation is an instance of DOMImplementation");
         ok(doc.implementation instanceof iframe.contentWindow.DOMImplementation, "created iframe doc.implementation is not an instance of iframe's DOMImplementation");
-        ok(Object.getPrototypeOf(doc) !== Object.getPrototypeOf(document), "created iframe doc's prototype same as doc prototype");
-        ok(Object.getPrototypeOf(doc) === iframe.contentWindow.HTMLDocument.prototype, "created iframe doc's prototype not iframe's HTMLDocument.prototype");
-
-        Object.defineProperty(doc, "winetest", { writable: true, enumerable: true, configurable: true, value: 42 });
-        test_own_data_prop_desc(doc, "winetest", true, true, true);
-
-        ok(Object.isFrozen(doc) === false, "created iframe doc isFrozen is not false");
-        ok(Object.isSealed(doc) === false, "created iframe doc isSealed is not false");
-        Object.freeze(doc);
-        ok(Object.isFrozen(doc) === true, "created iframe doc isFrozen is not true after freezing it");
-        ok(Object.isSealed(doc) === true, "created iframe doc isSealed is not true after freezing it");
 
         var r = Object.prototype.toString.call(iframe.contentWindow);
         ok(r === "[object Window]", "iframe's Window toString = " + r);
