--- a/dlls/win32u/message.c
+++ b/dlls/win32u/message.c
@@ -27,8 +27,6 @@
 #include <assert.h>
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
-#include "winternl.h"
-#include "ddk/wdm.h"
 #include "win32u_private.h"
 #include "ntuser_private.h"
 #include "winnls.h"
@@ -42,44 +40,6 @@
 WINE_DECLARE_DEBUG_CHANNEL(key);
 WINE_DECLARE_DEBUG_CHANNEL(relay);
 
-#define QS_DRIVER       0x80000000
-#define QS_HARDWARE     0x40000000
-#define QS_INTERNAL     (QS_DRIVER | QS_HARDWARE)
-
-static const struct _KUSER_SHARED_DATA *user_shared_data = (struct _KUSER_SHARED_DATA *)0x7ffe0000;
-
-static LONG atomic_load_long( const volatile LONG *ptr )
-{
-#if defined(__i386__) || defined(__x86_64__)
-    return *ptr;
-#else
-    return __atomic_load_n( ptr, __ATOMIC_SEQ_CST );
-#endif
-}
-
-static ULONG atomic_load_ulong( const volatile ULONG *ptr )
-{
-#if defined(__i386__) || defined(__x86_64__)
-    return *ptr;
-#else
-    return __atomic_load_n( ptr, __ATOMIC_SEQ_CST );
-#endif
-}
-
-static UINT64 get_tick_count(void)
-{
-    ULONG high, low;
-
-    do
-    {
-        high = atomic_load_long( &user_shared_data->TickCount.High1Time );
-        low = atomic_load_ulong( &user_shared_data->TickCount.LowPart );
-    }
-    while (high != atomic_load_long( &user_shared_data->TickCount.High2Time ));
-    /* note: we ignore TickCountMultiplier */
-    return (UINT64)high << 32 | low;
-}
-
 #define MAX_WINPROC_RECURSION  64
 
 #define WM_NCMOUSEFIRST WM_NCMOUSEMOVE
@@ -87,16 +47,6 @@
 
 #define MAX_PACK_COUNT 4
 
-struct peek_message_filter
-{
-    HWND hwnd;
-    UINT first;
-    UINT last;
-    UINT mask;
-    UINT flags;
-    BOOL internal;
-};
-
 /* info about the message currently being received by the current thread */
 struct received_message_info
 {
@@ -440,9 +390,13 @@
 
     user_check_not_lock();
 
-    if (!is_current_thread_window( hwnd )) return FALSE;
     if (!(win = get_win_ptr( hwnd ))) return FALSE;
     if (win == WND_OTHER_PROCESS || win == WND_DESKTOP) return FALSE;
+    if (win->tid != GetCurrentThreadId())
+    {
+        release_win_ptr( win );
+        return FALSE;
+    }
     params->func = win->winproc;
     params->ansi_dst = !(win->flags & WIN_ISUNICODE);
     is_dialog = win->dlgInfo != NULL;
@@ -466,15 +420,12 @@
     LRESULT result = 0;
     void *ret_ptr;
     ULONG ret_len;
-    NTSTATUS status;
 
     if (thread_info->recursion_count > MAX_WINPROC_RECURSION) return 0;
     thread_info->recursion_count++;
-    status = KeUserModeCallback( NtUserCallWinProc, params, size, &ret_ptr, &ret_len );
+    KeUserModeCallback( NtUserCallWinProc, params, size, &ret_ptr, &ret_len );
     thread_info->recursion_count--;
 
-    if (status) return result;
-
     if (ret_len >= sizeof(result))
     {
         result = *(LRESULT *)ret_ptr;
@@ -773,9 +724,11 @@
         minsize = sizeof(SCROLLINFO);
         break;
     case SBM_GETSCROLLINFO:
+    case WM_WINE_GETSCROLLINFO:
         if (!get_buffer_space( buffer, sizeof(SCROLLINFO), buffer_size )) return FALSE;
         break;
     case SBM_GETSCROLLBARINFO:
+    case WM_WINE_GETSCROLLBARINFO:
         if (!get_buffer_space( buffer, sizeof(SCROLLBARINFO), buffer_size )) return FALSE;
         break;
     case EM_GETSEL:
@@ -1171,9 +1124,11 @@
         push_data( data, (SCROLLINFO *)lparam, sizeof(SCROLLINFO) );
         return 0;
     case SBM_GETSCROLLINFO:
+    case WM_WINE_GETSCROLLINFO:
         push_data( data, (SCROLLINFO *)lparam, sizeof(SCROLLINFO) );
         return sizeof(SCROLLINFO);
     case SBM_GETSCROLLBARINFO:
+    case WM_WINE_GETSCROLLBARINFO:
     {
         const SCROLLBARINFO *info = (const SCROLLBARINFO *)lparam;
         size_t size = min( info->cbSize, sizeof(SCROLLBARINFO) );
@@ -1408,8 +1363,13 @@
         break;
     }
     case SBM_GETSCROLLINFO:
+    case WM_WINE_GETSCROLLINFO:
         push_data( data, (SCROLLINFO *)lparam, sizeof(SCROLLINFO) );
         break;
+    case SBM_GETSCROLLBARINFO:
+    case WM_WINE_GETSCROLLBARINFO:
+        push_data( data, (SCROLLBARINFO *)lparam, sizeof(SCROLLBARINFO) );
+        break;
     case EM_GETRECT:
     case LB_GETITEMRECT:
     case CB_GETDROPPEDCONTROLRECT:
@@ -1548,9 +1508,11 @@
         }
         break;
     case SBM_GETSCROLLINFO:
+    case WM_WINE_GETSCROLLINFO:
         memcpy( (SCROLLINFO *)lparam, buffer, min( sizeof(SCROLLINFO), size ));
         break;
     case SBM_GETSCROLLBARINFO:
+    case WM_WINE_GETSCROLLBARINFO:
         memcpy( (SCROLLBARINFO *)lparam, buffer, min( sizeof(SCROLLBARINFO), size ));
         break;
     case EM_GETRECT:
@@ -1737,11 +1699,7 @@
     case WM_COPYDATA:
     {
         const COPYDATASTRUCT *cds = lparam_ptr;
-        /* If cbData <= 2048 bytes, pack the data at the end of the message. Otherwise, pack the data
-         * in an extra user buffer to avoid potential stack overflows when calling KeUserModeCallback()
-         * because cbData can be very large. Manual tests of KiUserCallbackDispatcher() when receiving
-         * WM_COPYDATA messages show that Windows does similar things */
-        size = cds->cbData <= 2048 ? sizeof(*cds) + cds->cbData : sizeof(*cds);
+        size = sizeof(*cds) + cds->cbData;
         break;
     }
     case WM_HELP:
@@ -1756,9 +1714,11 @@
         break;
     case SBM_SETSCROLLINFO:
     case SBM_GETSCROLLINFO:
+    case WM_WINE_GETSCROLLINFO:
         size = sizeof(SCROLLINFO);
         break;
     case SBM_GETSCROLLBARINFO:
+    case WM_WINE_GETSCROLLBARINFO:
         size = sizeof(SCROLLBARINFO);
         break;
     case EM_GETSEL:
@@ -1830,20 +1790,15 @@
  *           pack_user_message
  *
  * Copy message to a buffer for passing to client.
- *
- * ret_extra_buffer returns an extra user buffer allocated to store large message data, for example,
- * for WM_COPYDATA. Call NtFreeVirtualMemory() for *ret_extra_buffer when done using it if it's not NULL.
  */
 void pack_user_message( void *buffer, size_t size, UINT message,
-                        WPARAM wparam, LPARAM lparam, BOOL ansi, void **ret_extra_buffer )
+                        WPARAM wparam, LPARAM lparam, BOOL ansi )
 {
     const void *lparam_ptr = (const void *)lparam;
     void const *inline_ptr = (void *)0xffffffff;
 
     if (!size) return;
 
-    *ret_extra_buffer = NULL;
-
     switch (message)
     {
     case WM_NCCREATE:
@@ -1880,47 +1835,9 @@
     case WM_COPYDATA:
     {
         const COPYDATASTRUCT *cds = lparam_ptr;
-
-        if (!cds->lpData)
-        {
-            size = sizeof(*cds);
-        }
-        else
-        {
-            /* If cbData <= 2048 bytes, pack the data at the end of the message. Otherwise, pack the data
-             * in an extra user buffer to avoid potential stack overflow when calling KeUserModeCallback()
-             * because cbData can be very large. Manual tests of KiUserCallbackDispatcher() when receiving
-             * WM_COPYDATA messages show that Windows does similar things */
-            if (cds->cbData <= 2048)
-            {
-                memcpy( (char *)buffer + sizeof(*cds), cds->lpData, cds->cbData );
-                size = sizeof(*cds);
-            }
-            else
-            {
-                COPYDATASTRUCT *tmp_cds = buffer;
-                SIZE_T extra_buffer_size;
-                unsigned int status;
-
-                memcpy( tmp_cds, cds, sizeof(*cds) );
-
-                extra_buffer_size = cds->cbData;
-                status = NtAllocateVirtualMemory( GetCurrentProcess(), ret_extra_buffer, zero_bits,
-                                                  &extra_buffer_size, MEM_RESERVE | MEM_COMMIT,
-                                                  PAGE_READWRITE );
-                if (!status)
-                {
-                    memcpy( *ret_extra_buffer, cds->lpData, cds->cbData );
-                    tmp_cds->lpData = *ret_extra_buffer;
-                }
-                else
-                {
-                    tmp_cds->lpData = NULL;
-                    tmp_cds->cbData = 0;
-                }
-                return;
-            }
-        }
+        if (cds->lpData && cds->cbData)
+            memcpy( (char *)buffer + sizeof(*cds), cds->lpData, cds->cbData );
+        size = sizeof(*cds);
         break;
     }
     case EM_GETSEL:
@@ -2041,9 +1958,11 @@
         break;
     case SBM_SETSCROLLINFO:
     case SBM_GETSCROLLINFO:
+    case WM_WINE_GETSCROLLINFO:
         copy_size = sizeof(SCROLLINFO);
         break;
     case SBM_GETSCROLLBARINFO:
+    case WM_WINE_GETSCROLLBARINFO:
         copy_size = sizeof(SCROLLBARINFO);
         break;
     case EM_GETSEL:
@@ -2130,11 +2049,11 @@
 }
 
 /***********************************************************************
- *           NtUserReplyMessage  (win32u.@)
+ *           reply_message_result
  *
  * Send a reply to a sent message and update thread receive info.
  */
-BOOL WINAPI NtUserReplyMessage( LRESULT result )
+BOOL reply_message_result( LRESULT result )
 {
     struct user_thread_info *thread_info = get_user_thread_info();
     struct received_message_info *info = thread_info->receive_info;
@@ -2192,12 +2111,12 @@
         return set_window_long( hwnd, (short)LOWORD(wparam), HIWORD(wparam), lparam, FALSE );
     case WM_WINE_SETSTYLE:
         if (is_desktop_window( hwnd )) return 0;
-        return set_window_style_bits( hwnd, wparam, lparam );
+        return set_window_style( hwnd, wparam, lparam );
     case WM_WINE_SETACTIVEWINDOW:
     {
         HWND prev;
 
-        if (!wparam && NtUserGetForegroundWindow() == hwnd) return 0;
+        if (!wparam && NtUserGetWindowThread( NtUserGetForegroundWindow(), NULL ) == GetCurrentThreadId()) return 0;
         if (!set_active_window( (HWND)wparam, &prev, FALSE, TRUE, lparam )) return 0;
         return (LRESULT)prev;
     }
@@ -2224,37 +2143,36 @@
     }
     case WM_WINE_WINDOW_STATE_CHANGED:
     {
-        UINT state_cmd, swp_flags;
+        UINT state_cmd, config_cmd;
         RECT window_rect;
         HWND foreground;
 
-        if (!user_driver->pGetWindowStateUpdates( hwnd, &state_cmd, &swp_flags, &window_rect, &foreground )) return 0;
-        window_rect = map_rect_raw_to_virt( window_rect, get_thread_dpi() );
-
-        if (foreground) set_foreground_window( foreground, FALSE, TRUE );
-        switch (state_cmd)
+        if (!user_driver->pGetWindowStateUpdates( hwnd, &state_cmd, &config_cmd, &window_rect, &foreground )) return 0;
+        if (foreground) NtUserSetForegroundWindow( foreground );
+        if (state_cmd)
+        {
+            if (LOWORD(state_cmd) == SC_RESTORE && HIWORD(state_cmd)) NtUserSetForegroundWindow( hwnd );
+            send_message( hwnd, WM_SYSCOMMAND, LOWORD(state_cmd), 0 );
+
+            /* state change might have changed the window config already, check again */
+            user_driver->pGetWindowStateUpdates( hwnd, &state_cmd, &config_cmd, &window_rect, &foreground );
+            if (foreground) NtUserSetForegroundWindow( foreground );
+            if (state_cmd) WARN( "window %p state needs another update, ignoring\n", hwnd );
+        }
+        if (config_cmd)
         {
-        case SC_RESTORE:
-            NtUserSetInternalWindowPos( hwnd, SW_SHOW, &window_rect, NULL );
-            /* fallthrough */
-        default:
-            send_message( hwnd, WM_SYSCOMMAND, state_cmd, 0 );
-            break;
-        case 0:
-            if (!swp_flags) break;
-            NtUserSetWindowPos( hwnd, 0, window_rect.left, window_rect.top, window_rect.right - window_rect.left,
-                                window_rect.bottom - window_rect.top, swp_flags );
-            break;
+            if (LOWORD(config_cmd) == SC_MOVE) NtUserSetRawWindowPos( hwnd, window_rect, HIWORD(config_cmd), FALSE );
+            else send_message( hwnd, WM_SYSCOMMAND, LOWORD(config_cmd), 0 );
         }
-
         return 0;
     }
     case WM_WINE_UPDATEWINDOWSTATE:
         update_window_state( hwnd );
         return 0;
-    case WM_WINE_SETPIXELFORMAT:
-        set_window_pixel_format( hwnd, wparam, lparam );
-        return 0;
+    case WM_WINE_GETSCROLLBARINFO:
+        return get_scroll_bar_info( hwnd, (LONG)wparam, (SCROLLBARINFO *)lparam );
+    case WM_WINE_GETSCROLLINFO:
+        return get_scroll_info( hwnd, (int)wparam, (SCROLLINFO *)lparam );
     case WM_WINE_TRACKMOUSEEVENT:
     {
         TRACKMOUSEEVENT info;
@@ -2330,11 +2248,10 @@
     struct win_proc_params p, *params = &p;
     BOOL ansi = ansi_dst && type == MSG_ASCII;
     size_t packed_size = 0, offset = sizeof(*params), reply_size;
-    void *ret_ptr, *extra_buffer = NULL;
-    SIZE_T extra_buffer_size = 0;
     LRESULT result = 0;
     CWPSTRUCT cwp;
     CWPRETSTRUCT cwpret;
+    void *ret_ptr;
     size_t ret_len = 0;
 
     if (msg & 0x80000000)
@@ -2366,12 +2283,10 @@
 
     if (type == MSG_OTHER_PROCESS) params->ansi = FALSE;
     if (packed_size)
-        pack_user_message( (char *)params + offset, packed_size, msg, wparam, lparam, ansi, &extra_buffer );
+        pack_user_message( (char *)params + offset, packed_size, msg, wparam, lparam, ansi );
 
     result = dispatch_win_proc_params( params, offset + packed_size, &ret_ptr, &ret_len );
     if (params != &p) free( params );
-    if (extra_buffer)
-        NtFreeVirtualMemory( GetCurrentProcess(), &extra_buffer, &extra_buffer_size, MEM_RELEASE );
 
     copy_user_result( ret_ptr, min( ret_len, reply_size ), result, msg, wparam, lparam, ansi );
 
@@ -2464,6 +2379,12 @@
  */
 static BOOL process_pointer_message( MSG *msg, UINT hw_id, const struct hardware_msg_data *msg_data )
 {
+    RECT rect;
+
+    SetRect( &rect, LOWORD(msg->lParam), HIWORD(msg->lParam), LOWORD(msg->lParam), HIWORD(msg->lParam) );
+    rect = map_rect_raw_to_virt( rect, get_thread_dpi() );
+    msg->lParam = MAKELPARAM(rect.left, rect.top);
+
     msg->pt = point_phys_to_win_dpi( msg->hwnd, msg->pt );
     return TRUE;
 }
@@ -2552,6 +2473,33 @@
     return TRUE;
 }
 
+
+static WORD pointer_buttons_from_mouse_buttons( WORD mouse_flags )
+{
+    static const struct
+    {
+        WORD mouse_flag;
+        WORD pointer_flag;
+    }
+    flags[] =
+    {
+        { MK_LBUTTON, POINTER_MESSAGE_FLAG_FIRSTBUTTON },
+        { MK_RBUTTON, POINTER_MESSAGE_FLAG_SECONDBUTTON },
+        { MK_MBUTTON, POINTER_MESSAGE_FLAG_THIRDBUTTON },
+        { MK_XBUTTON1, POINTER_MESSAGE_FLAG_FOURTHBUTTON },
+        { MK_XBUTTON2, POINTER_MESSAGE_FLAG_FIFTHBUTTON },
+    };
+
+    WORD pointer_flags = 0;
+    unsigned int i;
+
+    for (i = 0; i < ARRAY_SIZE(flags); ++i)
+    {
+        if (mouse_flags & flags[i].mouse_flag) pointer_flags |= flags[i].pointer_flag;
+    }
+    return pointer_flags;
+}
+
 /***********************************************************************
  *          process_mouse_message
  *
@@ -2605,6 +2553,56 @@
     msg->pt = point_phys_to_win_dpi( msg->hwnd, msg->pt );
     set_thread_dpi_awareness_context( get_window_dpi_awareness_context( msg->hwnd ));
 
+    if ((extra_info & 0xffffff00) != 0xff515700 && enable_mouse_in_pointer)
+    {
+        WORD flags = POINTER_MESSAGE_FLAG_INRANGE, pointer_button_flags;
+        DWORD message = 0;
+
+        pointer_button_flags = pointer_buttons_from_mouse_buttons( LOWORD( msg->wParam ));
+        if (pointer_button_flags) flags |= pointer_button_flags | POINTER_MESSAGE_FLAG_INCONTACT;
+
+        switch (msg->message)
+        {
+        case WM_MOUSEMOVE:
+            message = WM_POINTERUPDATE;
+            if (!pointer_button_flags) flags |= POINTER_MESSAGE_FLAG_PRIMARY;
+            break;
+        case WM_LBUTTONDOWN:
+        case WM_RBUTTONDOWN:
+        case WM_MBUTTONDOWN:
+        case WM_XBUTTONDOWN:
+            if (pointer_button_flags & (pointer_button_flags - 1))
+            {
+                /* More than one flag in pointer_button_flags, some buttons were already pressed.
+                 * WM_POINTERDOWN is only sent on the first button press. */
+                message = WM_POINTERUPDATE;
+            }
+            else
+            {
+                message = WM_POINTERDOWN;
+                flags |= POINTER_MESSAGE_FLAG_PRIMARY;
+            }
+            break;
+        case WM_LBUTTONUP:
+        case WM_RBUTTONUP:
+        case WM_MBUTTONUP:
+        case WM_XBUTTONUP:
+            /* WM_POINTERUP is only sent once all the buttons are up. */
+            message = pointer_button_flags ? WM_POINTERUPDATE : WM_POINTERUP;
+            break;
+        case WM_MOUSEWHEEL:
+            message = WM_POINTERWHEEL;
+            flags = HIWORD( msg->wParam );
+            break;
+        case WM_MOUSEHWHEEL:
+            message = WM_POINTERHWHEEL;
+            flags = HIWORD( msg->wParam );
+            break;
+        }
+
+        if (message) send_message( msg->hwnd, message, MAKELONG( 1, flags ), MAKELONG( msg->pt.x, msg->pt.y ) );
+    }
+
     /* FIXME: is this really the right place for this hook? */
     event.message = msg->message;
     event.time    = msg->time;
@@ -2747,7 +2745,7 @@
                     /* fall through */
                 case MA_ACTIVATE:
                 case 0:
-                    if (!set_foreground_window( hwndTop, TRUE, FALSE )) eat_msg = TRUE;
+                    if (!set_foreground_window( hwndTop, TRUE )) eat_msg = TRUE;
                     break;
                 default:
                     WARN( "unknown WM_MOUSEACTIVATE code %d\n", ret );
@@ -2814,7 +2812,7 @@
  * returns FALSE if we need to make a server request to update the queue masks or bits
  */
 static BOOL check_queue_bits( UINT wake_mask, UINT changed_mask, UINT signal_bits, UINT clear_bits,
-                              UINT *wake_bits, UINT *changed_bits, BOOL internal )
+                              UINT *wake_bits, UINT *changed_bits )
 {
     struct object_lock lock = OBJECT_LOCK_INIT;
     const queue_shm_t *queue_shm;
@@ -2823,9 +2821,8 @@
 
     while ((status = get_shared_queue( &lock, &queue_shm )) == STATUS_PENDING)
     {
-        if (internal) skip = !(queue_shm->internal_bits & QS_HARDWARE);
         /* if the masks need an update */
-        else if (queue_shm->wake_mask != wake_mask) skip = FALSE;
+        if (queue_shm->wake_mask != wake_mask) skip = FALSE;
         else if (queue_shm->changed_mask != changed_mask) skip = FALSE;
         /* or if some bits need to be cleared, or queue is signaled */
         else if (queue_shm->wake_bits & signal_bits) skip = FALSE;
@@ -2834,7 +2831,7 @@
         {
             *wake_bits = queue_shm->wake_bits;
             *changed_bits = queue_shm->changed_bits;
-            skip = get_tick_count() - (UINT64)queue_shm->access_time / 10000 < 3000; /* avoid hung queue */
+            skip = TRUE;
         }
     }
 
@@ -2849,14 +2846,13 @@
  * available; -1 on error.
  * All pending sent messages are processed before returning.
  */
-static int peek_message( MSG *msg, const struct peek_message_filter *filter )
+int peek_message( MSG *msg, const struct peek_message_filter *filter )
 {
     LRESULT result;
     HWND hwnd = filter->hwnd;
     UINT first = filter->first, last = filter->last, flags = filter->flags;
     struct user_thread_info *thread_info = get_user_thread_info();
     INPUT_MESSAGE_SOURCE prev_source = thread_info->client_info.msg_source;
-    HANDLE idle_event = thread_info->idle_event;
     struct received_message_info info;
     unsigned int hw_id = 0;  /* id of previous hardware message */
     unsigned char buffer_init[1024];
@@ -2890,8 +2886,9 @@
         thread_info->client_info.msg_source = prev_source;
         wake_mask = filter->mask & (QS_SENDMESSAGE | QS_SMRESULT);
 
-        if (check_queue_bits( wake_mask, filter->mask, wake_mask | signal_bits, filter->mask | clear_bits,
-                              &wake_bits, &changed_bits, filter->internal ))
+        if (!filter->waited && NtGetTickCount() - thread_info->last_getmsg_time < 3000 && /* avoid hung queue */
+            check_queue_bits( wake_mask, filter->mask, wake_mask | signal_bits, filter->mask | clear_bits,
+                              &wake_bits, &changed_bits ))
             res = STATUS_PENDING;
         else SERVER_START_REQ( get_message )
         {
@@ -2904,6 +2901,7 @@
             req->wake_mask = wake_mask;
             req->changed_mask = filter->mask;
             wine_server_set_reply( req, buffer, buffer_size );
+            thread_info->last_getmsg_time = NtGetTickCount();
             if (!(res = wine_server_call( req )))
             {
                 size = wine_server_reply_size( reply );
@@ -2924,11 +2922,7 @@
         if (res)
         {
             if (buffer != buffer_init) free( buffer );
-            if (res == STATUS_PENDING)
-            {
-                if (hwnd == (HWND)-1 && idle_event) NtSetEvent( idle_event, NULL );
-                return 0;
-            }
+            if (res == STATUS_PENDING) return 0;
             if (res != STATUS_BUFFER_OVERFLOW)
             {
                 RtlSetLastWin32Error( RtlNtStatusToDosError(res) );
@@ -3097,6 +3091,7 @@
                         .last = info.msg.message,
                         .mask = filter->mask,
                         .internal = filter->internal,
+                        .waited = TRUE,
                     };
                     peek_message( msg, &new_filter );
                 }
@@ -3126,9 +3121,14 @@
                     info.msg.lParam = result->lparam;
                 }
             }
-            if (info.msg.message == WM_TIMER || info.msg.message == WM_SYSTIMER)
+            /* CXHACK 19488 */
+            if (info.msg.message == WM_PAINT &&
+                    flags == (PM_REMOVE | PM_QS_INPUT | PM_QS_POSTMESSAGE | PM_QS_PAINT | PM_QS_SENDMESSAGE) &&
+                    (get_window_long( info.msg.hwnd, GWL_EXSTYLE ) & WS_EX_COMPOSITED ))
             {
-                if (!(flags & PM_NOYIELD) && idle_event) NtSetEvent( idle_event, NULL );
+                send_message( info.msg.hwnd, info.msg.message, info.msg.wParam, info.msg.lParam );
+                flags &= ~PM_QS_PAINT;
+                continue;
             }
             *msg = info.msg;
             msg->pt = point_phys_to_win_dpi( info.msg.hwnd, info.msg.pt );
@@ -3185,95 +3185,15 @@
         SERVER_START_REQ( get_msg_queue_handle )
         {
             wine_server_call( req );
-            thread_info->server_queue = wine_server_ptr_handle( reply->handle );
-            thread_info->idle_event = wine_server_ptr_handle( reply->idle_event );
+            ret = wine_server_ptr_handle( reply->handle );
         }
         SERVER_END_REQ;
-        if (!(ret = thread_info->server_queue)) ERR( "Cannot get server thread queue\n" );
+        thread_info->server_queue = ret;
+        if (!ret) ERR( "Cannot get server thread queue\n" );
     }
     return ret;
 }
 
-static BOOL is_queue_signaled(void)
-{
-    struct object_lock lock = OBJECT_LOCK_INIT;
-    const queue_shm_t *queue_shm;
-    BOOL signaled = FALSE;
-    UINT status;
-
-    while ((status = get_shared_queue( &lock, &queue_shm )) == STATUS_PENDING)
-        signaled = (queue_shm->wake_bits & queue_shm->wake_mask) ||
-                   (queue_shm->changed_bits & queue_shm->changed_mask);
-    if (status) return FALSE;
-
-    return signaled;
-}
-
-static BOOL check_queue_masks( UINT wake_mask, UINT changed_mask )
-{
-    struct object_lock lock = OBJECT_LOCK_INIT;
-    const queue_shm_t *queue_shm;
-    BOOL skip = FALSE;
-    UINT status;
-
-    while ((status = get_shared_queue( &lock, &queue_shm )) == STATUS_PENDING)
-    {
-        if (queue_shm->wake_mask != wake_mask || queue_shm->changed_mask != changed_mask) skip = FALSE;
-        else skip = get_tick_count() - (UINT64)queue_shm->access_time / 10000 < 3000; /* avoid hung queue */
-    }
-
-    if (status) return FALSE;
-    return skip;
-}
-
-static BOOL check_internal_bits( UINT mask )
-{
-    struct object_lock lock = OBJECT_LOCK_INIT;
-    const queue_shm_t *queue_shm;
-    BOOL signaled = FALSE;
-    UINT status;
-
-    while ((status = get_shared_queue( &lock, &queue_shm )) == STATUS_PENDING)
-        signaled = queue_shm->internal_bits & mask;
-    if (status) return FALSE;
-
-    return signaled;
-}
-
-static BOOL process_driver_events( UINT events_mask, UINT wake_mask, UINT changed_mask )
-{
-    BOOL drained = FALSE;
-
-    if (check_internal_bits( QS_DRIVER )) drained = user_driver->pProcessEvents( events_mask );
-
-    if (drained || !check_queue_masks( wake_mask, changed_mask ))
-    {
-        SERVER_START_REQ( set_queue_mask )
-        {
-            req->poll_events = drained;
-            req->wake_mask = wake_mask;
-            req->changed_mask = changed_mask;
-            wine_server_call( req );
-        }
-        SERVER_END_REQ;
-    }
-
-    /* process every pending internal hardware messages */
-    if (check_internal_bits( QS_HARDWARE ))
-    {
-        struct peek_message_filter filter = {.internal = TRUE};
-        MSG msg;
-        peek_message( &msg, &filter );
-    }
-
-    return is_queue_signaled();
-}
-
-void check_for_events( UINT flags )
-{
-    if (!process_driver_events( flags, 0, 0 ) && !(flags & QS_PAINT)) flush_window_surfaces( TRUE );
-}
-
 /* monotonic timer tick for throttling driver event checks */
 static inline LONGLONG get_driver_check_time(void)
 {
@@ -3288,7 +3208,7 @@
     if (get_user_thread_info()->last_driver_time != get_driver_check_time())
     {
         flush_window_surfaces( FALSE );
-        process_driver_events( QS_ALLINPUT, 0, 0 );
+        user_driver->pProcessEvents( QS_ALLINPUT );
         get_user_thread_info()->last_driver_time = get_driver_check_time();
     }
 }
@@ -3302,21 +3222,13 @@
 }
 
 /* wait for message or signaled handle */
-static DWORD wait_message( DWORD count, const HANDLE *handles, DWORD timeout, DWORD wake_mask, DWORD changed_mask, DWORD flags )
+static DWORD wait_message( DWORD count, const HANDLE *handles, DWORD timeout, DWORD mask, DWORD flags )
 {
     struct thunk_lock_params params = {.dispatch.callback = thunk_lock_callback};
-    WAIT_TYPE type = flags & MWMO_WAITALL ? WaitAll : WaitAny;
-    LARGE_INTEGER time, now, *abs;
+    LARGE_INTEGER time;
+    DWORD ret;
     void *ret_ptr;
     ULONG ret_len;
-    HANDLE event;
-    DWORD ret;
-
-    if ((abs = get_nt_timeout( &time, timeout )))
-    {
-        NtQuerySystemTime( &now );
-        abs->QuadPart = now.QuadPart - abs->QuadPart;
-    }
 
     if (!KeUserDispatchCallback( &params.dispatch, sizeof(params), &ret_ptr, &ret_len ) &&
         ret_len == sizeof(params.locks))
@@ -3325,16 +3237,17 @@
         params.restore = TRUE;
     }
 
-    process_driver_events( QS_ALLINPUT, wake_mask, changed_mask );
-    if (!(changed_mask & QS_SMRESULT) && (event = get_user_thread_info()->idle_event)) NtSetEvent( event, NULL );
-
-    do ret = NtWaitForMultipleObjects( count, handles, type, !!(flags & MWMO_ALERTABLE), abs );
-    while (ret == count - 1 && !process_driver_events( QS_ALLINPUT, wake_mask, changed_mask ));
-
-    if (HIWORD(ret)) /* is it an error code? */
+    if (user_driver->pProcessEvents( mask )) ret = count - 1;
+    else
     {
-        RtlSetLastWin32Error( RtlNtStatusToDosError(ret) );
-        ret = WAIT_FAILED;
+        ret = NtWaitForMultipleObjects( count, handles, !(flags & MWMO_WAITALL),
+                                        !!(flags & MWMO_ALERTABLE), get_nt_timeout( &time, timeout ));
+        if (ret == count - 1) user_driver->pProcessEvents( mask );
+        else if (HIWORD(ret)) /* is it an error code? */
+        {
+            RtlSetLastWin32Error( RtlNtStatusToDosError(ret) );
+            ret = WAIT_FAILED;
+        }
     }
 
     if (ret == WAIT_TIMEOUT && !count && !timeout) NtYieldExecution();
@@ -3346,6 +3259,23 @@
 }
 
 /***********************************************************************
+ *           check_queue_masks
+ */
+static BOOL check_queue_masks( UINT wake_mask, UINT changed_mask )
+{
+    struct object_lock lock = OBJECT_LOCK_INIT;
+    const queue_shm_t *queue_shm;
+    BOOL skip = FALSE;
+    UINT status;
+
+    while ((status = get_shared_queue( &lock, &queue_shm )) == STATUS_PENDING)
+        skip = queue_shm->wake_mask == wake_mask && queue_shm->changed_mask == changed_mask;
+
+    if (status) return FALSE;
+    return skip;
+}
+
+/***********************************************************************
  *           wait_objects
  *
  * Wait for multiple objects including the server queue, with specific queue masks.
@@ -3357,7 +3287,19 @@
 
     flush_window_surfaces( TRUE );
 
-    return wait_message( count, handles, timeout, wake_mask, changed_mask, flags );
+    if (!check_queue_masks( wake_mask, changed_mask ))
+    {
+        SERVER_START_REQ( set_queue_mask )
+        {
+            req->wake_mask    = wake_mask;
+            req->changed_mask = changed_mask;
+            req->skip_wait    = 0;
+            wine_server_call( req );
+        }
+        SERVER_END_REQ;
+    }
+
+    return wait_message( count, handles, timeout, changed_mask, flags );
 }
 
 static HANDLE normalize_std_handle( HANDLE handle )
@@ -3530,6 +3472,7 @@
 
     filter.mask = mask;
     filter.flags = PM_REMOVE | (mask << 16);
+    filter.waited = TRUE;
     while (!(ret = peek_message( msg, &filter )))
     {
         wait_objects( 1, &server_queue, INFINITE, mask & (QS_SENDMESSAGE | QS_SMRESULT), mask, 0 );
@@ -3666,12 +3609,13 @@
         UINT wake_bits, changed_bits;
 
         if (check_queue_bits( wake_mask, wake_mask, wake_mask, wake_mask,
-                              &wake_bits, &changed_bits, FALSE ))
+                              &wake_bits, &changed_bits ))
             wake_bits = wake_bits & wake_mask;
         else SERVER_START_REQ( set_queue_mask )
         {
             req->wake_mask    = wake_mask;
             req->changed_mask = wake_mask;
+            req->skip_wait    = 1;
             wine_server_call( req );
             wake_bits = reply->wake_bits & wake_mask;
         }
@@ -3685,7 +3629,7 @@
             continue;
         }
 
-        wait_message( 1, &server_queue, INFINITE, wake_mask, wake_mask, 0 );
+        wait_message( 1, &server_queue, INFINITE, wake_mask, 0 );
     }
 }
 
@@ -4119,7 +4063,8 @@
 {
     LRESULT ret = 0;
 
-    TRACE( "(hwnd=%p,msg=%s,wp=%#lx,lp=%#lx)\n", hwnd, debugstr_msg_name( msg, hwnd ), (long)wparam, lparam );
+    TRACE( "(hwnd=%p,msg=%s,wp=%#lx,lp=%#lx)\n", hwnd, debugstr_msg_name( msg, hwnd ),
+           (long)wparam, (long)lparam );
 
     switch(msg)
     {
@@ -4325,9 +4270,7 @@
     thread_info->msg_source = msg_source_unavailable;
     spy_enter_message( SPY_SENDMESSAGE, info->hwnd, info->msg, info->wparam, info->lparam );
 
-    if (info->dest_tid != GetCurrentThreadId() ||
-        /* When the callback pointer is NULL and the data is 1, don't wait for the result */
-        (info->type == MSG_CALLBACK && !info->callback && info->data == 1))
+    if (info->dest_tid != GetCurrentThreadId())
     {
         if (dest_pid != GetCurrentProcessId() && (info->type == MSG_ASCII || info->type == MSG_UNICODE))
             info->type = MSG_OTHER_PROCESS;
@@ -4434,10 +4377,8 @@
     return ret;
 }
 
-/***********************************************************************
- *           NtUserKillSystemTimer (win32u.@)
- */
-BOOL WINAPI NtUserKillSystemTimer( HWND hwnd, UINT_PTR id )
+/* see KillSystemTimer */
+BOOL kill_system_timer( HWND hwnd, UINT_PTR id )
 {
     BOOL ret;
 
@@ -4707,11 +4648,8 @@
     case NtUserPostDdeCall:
         return post_dde_message_call( hwnd, msg, wparam, lparam, result_info );
 
-    case NtUserWintabDriverCall:
-        return user_driver->pWintabProc( hwnd, msg, wparam, lparam, result_info );
-
     default:
-        FIXME( "%p %x %lx %lx %p %x %x\n", hwnd, msg, (long)wparam, lparam, result_info, type, ansi );
+        FIXME( "%p %x %lx %lx %p %x %x\n", hwnd, msg, (long)wparam, lparam, result_info, (int)type, ansi );
     }
     return 0;
 }
