diff --git a/dlls/kernel32/tests/sync.c b/dlls/kernel32/tests/sync.c
--- a/dlls/kernel32/tests/sync.c
+++ b/dlls/kernel32/tests/sync.c
@@ -56,12 +56,11 @@
 static BOOLEAN (WINAPI *pTryAcquireSRWLockExclusive)(PSRWLOCK);
 static BOOLEAN (WINAPI *pTryAcquireSRWLockShared)(PSRWLOCK);
 
-static DWORD (WINAPI *pQueueUserAPC2)(PAPCFUNC,HANDLE,ULONG_PTR,QUEUE_USER_APC_FLAGS);
-
 static NTSTATUS (WINAPI *pNtAllocateVirtualMemory)(HANDLE, PVOID *, ULONG_PTR, SIZE_T *, ULONG, ULONG);
 static NTSTATUS (WINAPI *pNtFreeVirtualMemory)(HANDLE, PVOID *, SIZE_T *, ULONG);
+static NTSTATUS (WINAPI *pNtQuerySystemTime)(LARGE_INTEGER *);
 static NTSTATUS (WINAPI *pNtWaitForSingleObject)(HANDLE, BOOLEAN, const LARGE_INTEGER *);
-static NTSTATUS (WINAPI *pNtWaitForMultipleObjects)(ULONG,const HANDLE*,WAIT_TYPE,BOOLEAN,const LARGE_INTEGER*);
+static NTSTATUS (WINAPI *pNtWaitForMultipleObjects)(ULONG,const HANDLE*,BOOLEAN,BOOLEAN,const LARGE_INTEGER*);
 static PSLIST_ENTRY (__fastcall *pRtlInterlockedPushListSList)(PSLIST_HEADER list, PSLIST_ENTRY first,
                                                                PSLIST_ENTRY last, ULONG count);
 static PSLIST_ENTRY (WINAPI *pRtlInterlockedPushListSListEx)(PSLIST_HEADER list, PSLIST_ENTRY first,
@@ -75,7 +74,7 @@
 
 #ifdef __i386__
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 struct fastcall_thunk
 {
     BYTE pop_edx;   /* popl %edx            (ret addr) */
@@ -84,7 +83,7 @@
     BYTE xchg[3];   /* xchgl (%esp),%edx    (param 2) */
     WORD jmp_eax;   /* jmp  *%eax */
 };
-#pragma pack(pop)
+#include "poppack.h"
 
 static void * (WINAPI *call_fastcall_func4)(void *func, const void *a, const void *b, const void *c, const void *d);
 
@@ -234,35 +233,27 @@
     ok(GetLastError() == ERROR_FILE_NOT_FOUND, "wrong error %lu\n", GetLastError());
 }
 
-struct test_mutex_thread_params
-{
-    HANDLE mutex;
-    HANDLE start_event;
-    HANDLE stop_event;
-    BOOL owner;
-};
-
-static DWORD WINAPI test_mutex_thread(void *arg)
-{
-    struct test_mutex_thread_params *params = arg;
+static HANDLE mutex, mutex2, mutices[2];
+
+static DWORD WINAPI mutex_thread( void *param )
+{
+    DWORD expect = (DWORD)(DWORD_PTR)param;
     DWORD ret;
 
-    ret = WaitForSingleObject(params->mutex, INFINITE);
-    if (params->owner) ok(!ret, "got %#lx\n", ret);
-    else ok(ret == WAIT_ABANDONED, "got %#lx\n", ret);
-    SetEvent(params->start_event);
-
-    ret = WaitForSingleObject(params->stop_event, INFINITE);
-    ok(!ret, "got %#lx\n", ret);
+    ret = WaitForSingleObject( mutex, 0 );
+    ok(ret == expect, "expected %lu, got %lu\n", expect, ret);
+
+    if (!ret) ReleaseMutex( mutex );
     return 0;
 }
 
 static void test_mutex(void)
 {
+    HANDLE thread;
     DWORD wait_ret;
     BOOL ret;
-    HANDLE hCreated, hOpened, owner_thread, waiter_thread;
-    struct test_mutex_thread_params params;
+    HANDLE hCreated;
+    HANDLE hOpened;
     int i;
     DWORD failed = 0;
 
@@ -277,7 +268,9 @@
 
     SetLastError(0xdeadbeef);
     hOpened = OpenMutexA(0, FALSE, "WineTestMutex");
+    todo_wine
     ok(hOpened == NULL, "OpenMutex succeeded\n");
+    todo_wine
     ok(GetLastError() == ERROR_ACCESS_DENIED, "wrong error %lu\n", GetLastError());
 
     SetLastError(0xdeadbeef);
@@ -296,7 +289,8 @@
     SetLastError(0xdeadbeef);
     hOpened = OpenMutexA(GENERIC_READ | GENERIC_WRITE, FALSE, "WineTestMutex");
     ok(hOpened != NULL, "OpenMutex failed with error %ld\n", GetLastError());
-    wait_ret = WaitForSingleObject(hOpened, INFINITE);
+    wait_ret = WaitForSingleObject(hOpened, 0);
+todo_wine_if(getenv("WINEESYNC"))   /* XFAIL: validation is not implemented */
     ok(wait_ret == WAIT_FAILED, "WaitForSingleObject succeeded\n");
     CloseHandle(hOpened);
 
@@ -316,7 +310,7 @@
             if ((1 << i) == ACCESS_SYSTEM_SECURITY)
                 todo_wine ok(GetLastError() == ERROR_PRIVILEGE_NOT_HELD, "wrong error %lu, access %x\n", GetLastError(), 1 << i);
             else
-                ok(GetLastError() == ERROR_ACCESS_DENIED, "wrong error %lu, , access %x\n", GetLastError(), 1 << i);
+                todo_wine ok(GetLastError() == ERROR_ACCESS_DENIED, "wrong error %lu, , access %x\n", GetLastError(), 1 << i);
             ReleaseMutex(hCreated);
             failed |=0x1 << i;
         }
@@ -327,6 +321,7 @@
 
     SetLastError(0xdeadbeef);
     ret = ReleaseMutex(hCreated);
+todo_wine_if(getenv("WINEESYNC"))   /* XFAIL: due to the above */
     ok(!ret && (GetLastError() == ERROR_NOT_OWNER),
         "ReleaseMutex should have failed with ERROR_NOT_OWNER instead of %ld\n", GetLastError());
 
@@ -366,41 +361,84 @@
 
     CloseHandle(hCreated);
 
-    params.start_event = CreateEventW(NULL, FALSE, FALSE, NULL);
-    params.stop_event = CreateEventW(NULL, FALSE, FALSE, NULL);
-    params.mutex = CreateMutexA(NULL, FALSE, NULL);
-
-    params.owner = TRUE;
-    owner_thread = CreateThread(NULL, 0, test_mutex_thread, &params, 0, NULL);
-    ok(!!owner_thread, "CreateThread failed, error %lu\n", GetLastError());
-    ret = WaitForSingleObject(params.start_event, 1000);
-    ok(!ret, "got %#x\n", ret);
-
-    params.owner = FALSE;
-    waiter_thread = CreateThread(NULL, 0, test_mutex_thread, &params, 0, NULL);
-    ok(!!waiter_thread, "CreateThread failed, error %lu\n", GetLastError());
-    ret = WaitForSingleObject(params.start_event, 100);
-    ok(ret == WAIT_TIMEOUT, "got %#x\n", ret);
-
-    CloseHandle(params.mutex);
-    ret = WaitForSingleObject(params.start_event, 100);
-    ok(ret == WAIT_TIMEOUT, "got %#x\n", ret);
-
-    TerminateThread(owner_thread, 0);
-    ret = WaitForSingleObject(owner_thread, 1000);
-    ok(!ret, "got %#x\n", ret);
-    ret = WaitForSingleObject(params.start_event, 1000);
-    ok(!ret, "got %#x\n", ret);
-
-    SetEvent(params.stop_event);
-    ret = WaitForSingleObject(waiter_thread, 1000);
-    ok(!ret, "got %#x\n", ret);
-
-    CloseHandle(owner_thread);
-    CloseHandle(waiter_thread);
-
-    CloseHandle(params.start_event);
-    CloseHandle(params.stop_event);
+    mutex = CreateMutexA( NULL, FALSE, NULL );
+    ok(!!mutex, "got error %lu\n", GetLastError());
+
+    ret = ReleaseMutex( mutex );
+    ok(!ret, "got %d\n", ret);
+    ok(GetLastError() == ERROR_NOT_OWNER, "got error %lu\n", GetLastError());
+
+    for (i = 0; i < 100; i++)
+    {
+        ret = WaitForSingleObject( mutex, 0 );
+        ok(ret == 0, "got %u\n", ret);
+    }
+
+    for (i = 0; i < 100; i++)
+    {
+        ret = ReleaseMutex( mutex );
+        ok(ret, "got error %lu\n", GetLastError());
+    }
+
+    ret = ReleaseMutex( mutex );
+    ok(!ret, "got %d\n", ret);
+    ok(GetLastError() == ERROR_NOT_OWNER, "got error %lu\n", GetLastError());
+
+    thread = CreateThread( NULL, 0, mutex_thread, (void *)0, 0, NULL );
+    ret = WaitForSingleObject( thread, 2000 );
+    ok(ret == 0, "wait failed: %u\n", ret);
+
+    WaitForSingleObject( mutex, 0 );
+
+    thread = CreateThread( NULL, 0, mutex_thread, (void *)WAIT_TIMEOUT, 0, NULL );
+    ret = WaitForSingleObject( thread, 2000 );
+    ok(ret == 0, "wait failed: %u\n", ret);
+
+    ret = ReleaseMutex( mutex );
+        ok(ret, "got error %lu\n", GetLastError());
+
+    thread = CreateThread( NULL, 0, mutex_thread, (void *)0, 0, NULL );
+    ret = WaitForSingleObject( thread, 2000 );
+    ok(ret == 0, "wait failed: %u\n", ret);
+
+    mutex2 = CreateMutexA( NULL, TRUE, NULL );
+    ok(!!mutex2, "got error %lu\n", GetLastError());
+
+    ret = ReleaseMutex( mutex2 );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = ReleaseMutex( mutex2 );
+    ok(!ret, "got %d\n", ret);
+    ok(GetLastError() == ERROR_NOT_OWNER, "got error %lu\n", GetLastError());
+
+    mutices[0] = mutex;
+    mutices[1] = mutex2;
+
+    ret = WaitForMultipleObjects( 2, mutices, FALSE, 0 );
+    ok(ret == 0, "got %u\n", ret);
+
+    ret = ReleaseMutex( mutex );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = ReleaseMutex( mutex2 );
+    ok(!ret, "got %d\n", ret);
+    ok(GetLastError() == ERROR_NOT_OWNER, "got error %lu\n", GetLastError());
+
+    ret = WaitForMultipleObjects( 2, mutices, TRUE, 0 );
+    ok(ret == 0, "got %u\n", ret);
+
+    ret = ReleaseMutex( mutex );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = ReleaseMutex( mutex2 );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = CloseHandle( mutex );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = CloseHandle( mutex2 );
+    ok(ret, "got error %lu\n", GetLastError());
+
 }
 
 static void test_slist(void)
@@ -576,12 +614,13 @@
 
 static void test_event(void)
 {
-    HANDLE handle, handle2;
+    HANDLE handle, handle2, handles[2];
     SECURITY_ATTRIBUTES sa;
     SECURITY_DESCRIPTOR sd;
     ACL acl;
     DWORD ret;
     BOOL val;
+    int i;
 
     /* no sd */
     handle = CreateEventA(NULL, FALSE, FALSE, __FILE__ ": Test Event");
@@ -685,11 +724,130 @@
     ok( ret, "QueryMemoryResourceNotification failed err %lu\n", GetLastError() );
     ok( val == FALSE || val == TRUE, "wrong value %u\n", val );
     CloseHandle( handle );
+
+    handle = CreateEventA( NULL, TRUE, FALSE, NULL );
+    ok(!!handle, "got error %lu\n", GetLastError());
+
+    ret = WaitForSingleObject( handle, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    ret = SetEvent( handle );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = SetEvent( handle );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    for (i = 0; i < 100; i++)
+    {
+        ret = WaitForSingleObject( handle, 0 );
+        ok(ret == 0, "got %lu\n", ret);
+    }
+
+    ret = ResetEvent( handle );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = ResetEvent( handle );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = WaitForSingleObject( handle, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    handle2 = CreateEventA( NULL, FALSE, TRUE, NULL );
+    ok(!!handle2, "got error %lu\n", GetLastError());
+
+    ret = WaitForSingleObject( handle2, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = WaitForSingleObject( handle2, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    ret = SetEvent( handle2 );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = SetEvent( handle2 );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = ResetEvent( handle2 );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = ResetEvent( handle2 );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = WaitForSingleObject( handle2, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    handles[0] = handle;
+    handles[1] = handle2;
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    SetEvent( handle );
+    SetEvent( handle2 );
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = WaitForSingleObject( handle2, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ResetEvent( handle );
+    SetEvent( handle2 );
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == 1, "got %lu\n", ret);
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    SetEvent( handle );
+    SetEvent( handle2 );
+
+    ret = WaitForMultipleObjects( 2, handles, TRUE, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = WaitForMultipleObjects( 2, handles, TRUE, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    SetEvent( handle2 );
+    ResetEvent( handle );
+
+    ret = WaitForMultipleObjects( 2, handles, TRUE, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    ret = WaitForSingleObject( handle2, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    handles[0] = handle2;
+    handles[1] = handle;
+    SetEvent( handle );
+    SetEvent( handle2 );
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == 1, "got %lu\n", ret);
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == 1, "got %lu\n", ret);
+
+    ret = CloseHandle( handle );
+    ok(ret, "got error %lu\n", GetLastError());
+
+    ret = CloseHandle( handle2 );
+    ok(ret, "got error %lu\n", GetLastError());
 }
 
 static void test_semaphore(void)
 {
-    HANDLE handle, handle2;
+    HANDLE handle, handle2, handles[2];
+    DWORD ret;
+    LONG prev;
+    int i;
 
     /* test case sensitivity */
 
@@ -731,6 +889,99 @@
     ok( GetLastError() == ERROR_INVALID_PARAMETER, "wrong error %lu\n", GetLastError());
 
     CloseHandle( handle );
+
+    handle = CreateSemaphoreA( NULL, 0, 5, NULL );
+    ok(!!handle, "CreateSemaphore failed: %lu\n", GetLastError());
+
+    ret = WaitForSingleObject( handle, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    ret = ReleaseSemaphore( handle, 1, &prev );
+    ok(ret, "got error %lu\n", GetLastError());
+    ok(prev == 0, "got prev %ld\n", prev);
+
+    ret = ReleaseSemaphore( handle, 1, &prev );
+    ok(ret, "got error %lu\n", GetLastError());
+    ok(prev == 1, "got prev %ld\n", prev);
+
+    ret = ReleaseSemaphore( handle, 5, &prev );
+    ok(!ret, "got %ld\n", ret);
+    ok(GetLastError() == ERROR_TOO_MANY_POSTS, "got error %lu\n", GetLastError());
+    ok(prev == 1, "got prev %ld\n", prev);
+
+    ret = ReleaseSemaphore( handle, 2, &prev );
+    ok(ret, "got error %lu\n", GetLastError());
+    ok(prev == 2, "got prev %ld\n", prev);
+
+    ret = ReleaseSemaphore( handle, 1, &prev );
+    ok(ret, "got error %lu\n", GetLastError());
+    ok(prev == 4, "got prev %ld\n", prev);
+
+    for (i = 0; i < 5; i++)
+    {
+        ret = WaitForSingleObject( handle, 0 );
+        ok(ret == 0, "got %lu\n", ret);
+    }
+
+    ret = WaitForSingleObject( handle, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    handle2 = CreateSemaphoreA( NULL, 3, 5, NULL );
+    ok(!!handle2, "CreateSemaphore failed: %lu\n", GetLastError());
+
+    ret = ReleaseSemaphore( handle2, 1, &prev );
+    ok(ret, "got error %lu\n", GetLastError());
+    ok(prev == 3, "got prev %ld\n", prev);
+
+    for (i = 0; i < 4; i++)
+    {
+        ret = WaitForSingleObject( handle2, 0 );
+        ok(ret == 0, "got %lu\n", ret);
+    }
+
+    ret = WaitForSingleObject( handle2, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    handles[0] = handle;
+    handles[1] = handle2;
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    ReleaseSemaphore( handle, 1, NULL );
+    ReleaseSemaphore( handle2, 1, NULL );
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == 1, "got %lu\n", ret);
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    ReleaseSemaphore( handle, 1, NULL );
+    ReleaseSemaphore( handle2, 1, NULL );
+
+    ret = WaitForMultipleObjects( 2, handles, TRUE, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = WaitForMultipleObjects( 2, handles, FALSE, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    ReleaseSemaphore( handle, 1, NULL );
+
+    ret = WaitForMultipleObjects( 2, handles, TRUE, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    ret = WaitForSingleObject( handle, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = CloseHandle( handle );
+    ok(ret, "got error %lu\n", ret);
+
+    ret = CloseHandle( handle2 );
+    ok(ret, "got error %lu\n", ret);
 }
 
 static void test_waitable_timer(void)
@@ -1285,30 +1536,16 @@
     return ULongToHandle(tmp);
 }
 
+#define TIMEOUT_INFINITE (((LONGLONG)0x7fffffff) << 32 | 0xffffffff)
+
 static void test_WaitForSingleObject(void)
 {
     HANDLE signaled, nonsignaled, invalid;
+    LARGE_INTEGER ntnow, ntthen;
     LARGE_INTEGER timeout;
     NTSTATUS status;
+    DWORD now, then;
     DWORD ret;
-    int i;
-    HANDLE waitable_pseudohandles[] =
-    {
-        NtCurrentProcess(),
-        NtCurrentThread(),
-    };
-    HANDLE non_waitable_pseudohandles[4];
-    HANDLE std_handles[] =
-    {
-        (HANDLE)STD_INPUT_HANDLE,
-        (HANDLE)STD_OUTPUT_HANDLE,
-        (HANDLE)STD_ERROR_HANDLE,
-    };
-
-    non_waitable_pseudohandles[0] = (HANDLE)~(ULONG_PTR)2;
-    non_waitable_pseudohandles[1] = GetCurrentProcessToken();
-    non_waitable_pseudohandles[2] = GetCurrentThreadToken();
-    non_waitable_pseudohandles[3] = GetCurrentThreadEffectiveToken();
 
     signaled = CreateEventW(NULL, TRUE, TRUE, NULL);
     nonsignaled = CreateEventW(NULL, TRUE, FALSE, NULL);
@@ -1377,44 +1614,82 @@
     ok(ret == WAIT_OBJECT_0, "expected WAIT_OBJECT_0, got %ld\n", ret);
     ok(GetLastError() == 0xdeadbeef, "expected 0xdeadbeef, got %ld\n", GetLastError());
 
-    /* waitable pseudo-handles are allowed in WaitForSingleObject and NtWaitForSingleObject,
-     * but not in  NtWaitForMultipleObjects, see test_WaitForMultipleObjects */
-    for (i = 0; i < ARRAY_SIZE(waitable_pseudohandles); i++)
-    {
-        ret = WaitForSingleObject(waitable_pseudohandles[i], 0);
-        ok(ret == WAIT_TIMEOUT, "expected WAIT_TIMEOUT, got %lu\n", ret);
-
-        timeout.QuadPart = 0;
-        status = pNtWaitForSingleObject(waitable_pseudohandles[i], FALSE, &timeout);
-        ok(status == STATUS_TIMEOUT, "expected STATUS_TIMEOUT, got %08lx\n", status);
-    }
-
-    /* non-waitable pseudo-handles return STATUS_INVALID_HANDLE */
-    for (i = 0; i < ARRAY_SIZE(non_waitable_pseudohandles); i++)
-    {
-        SetLastError(0xdeadbeef);
-        ret = WaitForSingleObject(non_waitable_pseudohandles[i], 0);
-        ok(ret == WAIT_FAILED, "expected WAIT_FAILED, got %ld\n", ret);
-        ok(GetLastError() == ERROR_INVALID_HANDLE, "expected ERROR_INVALID_HANDLE, got %ld\n", GetLastError());
-
-        status = pNtWaitForSingleObject(non_waitable_pseudohandles[i], FALSE, NULL);
-        todo_wine_if((non_waitable_pseudohandles[i] == GetCurrentProcessToken() ||
-                      non_waitable_pseudohandles[i] == GetCurrentThreadEffectiveToken()) &&
-                     NT_ERROR(status))
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-    }
-
-    /* std handles are only allowed in WaitForSingleObject and not NtWaitForSingleObject */
-    for (i = 0; i < ARRAY_SIZE(std_handles); i++)
-    {
-        ret = WaitForSingleObject(std_handles[i], 0);
-        ok(ret == WAIT_OBJECT_0 || ret == WAIT_TIMEOUT,
-           "expected WAIT_OBJECT_0 or WAIT_TIMEOUT), got %lu, GetLastError()=%lu\n",
-           ret, GetLastError());
-
-        status = pNtWaitForSingleObject(std_handles[i], FALSE, NULL);
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-    }
+    /* pseudo handles are allowed in WaitForSingleObject and NtWaitForSingleObject */
+    ret = WaitForSingleObject(GetCurrentProcess(), 100);
+    ok(ret == WAIT_TIMEOUT, "expected WAIT_TIMEOUT, got %lu\n", ret);
+
+    ret = WaitForSingleObject(GetCurrentThread(), 100);
+    ok(ret == WAIT_TIMEOUT, "expected WAIT_TIMEOUT, got %lu\n", ret);
+
+    timeout.QuadPart = -1000000;
+    status = pNtWaitForSingleObject(GetCurrentProcess(), FALSE, &timeout);
+    ok(status == STATUS_TIMEOUT, "expected STATUS_TIMEOUT, got %08lx\n", status);
+
+    timeout.QuadPart = -1000000;
+    status = pNtWaitForSingleObject(GetCurrentThread(), FALSE, &timeout);
+    ok(status == STATUS_TIMEOUT, "expected STATUS_TIMEOUT, got %08lx\n", status);
+
+    ret = WaitForSingleObject( signaled, 0 );
+    ok(ret == 0, "got %lu\n", ret);
+
+    ret = WaitForSingleObject( nonsignaled, 0 );
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+
+    /* test that a timed wait actually does wait */
+    now = GetTickCount();
+    ret = WaitForSingleObject( nonsignaled, 100 );
+    then = GetTickCount();
+    ok(ret == WAIT_TIMEOUT, "got %lu\n", ret);
+    ok(abs((then - now) - 100) < 5, "got %lu ms\n", then - now);
+
+    now = GetTickCount();
+    ret = WaitForSingleObject( signaled, 100 );
+    then = GetTickCount();
+    ok(ret == 0, "got %lu\n", ret);
+    ok(abs(then - now) < 5, "got %lu ms\n", then - now);
+
+    ret = WaitForSingleObject( signaled, INFINITE );
+    ok(ret == 0, "got %lu\n", ret);
+
+    /* test NT timeouts */
+    pNtQuerySystemTime( &ntnow );
+    timeout.QuadPart = ntnow.QuadPart + 100 * 10000;
+    status = pNtWaitForSingleObject( nonsignaled, FALSE, &timeout );
+    pNtQuerySystemTime( &ntthen );
+    ok(status == STATUS_TIMEOUT, "got %#lx\n", status);
+    ok(abs(((ntthen.QuadPart - ntnow.QuadPart) / 10000) - 100) < 5, "got %s ns\n",
+        wine_dbgstr_longlong((ntthen.QuadPart - ntnow.QuadPart) * 100));
+
+    pNtQuerySystemTime( &ntnow );
+    timeout.QuadPart = -100 * 10000;
+    status = pNtWaitForSingleObject( nonsignaled, FALSE, &timeout );
+    pNtQuerySystemTime( &ntthen );
+    ok(status == STATUS_TIMEOUT, "got %#lx\n", status);
+    ok(abs(((ntthen.QuadPart - ntnow.QuadPart) / 10000) - 100) < 5, "got %s ns\n",
+        wine_dbgstr_longlong((ntthen.QuadPart - ntnow.QuadPart) * 100));
+
+    status = pNtWaitForSingleObject( signaled, FALSE, NULL );
+    ok(status == 0, "got %#lx\n", status);
+
+    timeout.QuadPart = TIMEOUT_INFINITE;
+    status = pNtWaitForSingleObject( signaled, FALSE, &timeout );
+    ok(status == 0, "got %#lx\n", status);
+
+    pNtQuerySystemTime( &ntnow );
+    timeout.QuadPart = ntnow.QuadPart;
+    status = pNtWaitForSingleObject( nonsignaled, FALSE, &timeout );
+    pNtQuerySystemTime( &ntthen );
+    ok(status == STATUS_TIMEOUT, "got %#lx\n", status);
+    ok(abs((ntthen.QuadPart - ntnow.QuadPart) / 10000) < 5, "got %s ns\n",
+        wine_dbgstr_longlong((ntthen.QuadPart - ntnow.QuadPart) * 100));
+
+    pNtQuerySystemTime( &ntnow );
+    timeout.QuadPart = ntnow.QuadPart - 100 * 10000;
+    status = pNtWaitForSingleObject( nonsignaled, FALSE, &timeout );
+    pNtQuerySystemTime( &ntthen );
+    ok(status == STATUS_TIMEOUT, "got %#lx\n", status);
+    ok(abs((ntthen.QuadPart - ntnow.QuadPart) / 10000) < 5, "got %s ns\n",
+        wine_dbgstr_longlong((ntthen.QuadPart - ntnow.QuadPart) * 100));
 
     CloseHandle(signaled);
     CloseHandle(nonsignaled);
@@ -1422,28 +1697,15 @@
 
 static void test_WaitForMultipleObjects(void)
 {
+    LARGE_INTEGER timeout;
     NTSTATUS status;
     DWORD r;
     int i;
     HANDLE maxevents[MAXIMUM_WAIT_OBJECTS];
-    HANDLE pseudohandles[6];
-    HANDLE std_handles[] =
-    {
-        (HANDLE)STD_INPUT_HANDLE,
-        (HANDLE)STD_OUTPUT_HANDLE,
-        (HANDLE)STD_ERROR_HANDLE,
-    };
-
-    pseudohandles[0] = GetCurrentProcess();
-    pseudohandles[1] = GetCurrentThread();
-    pseudohandles[2] = (HANDLE)~(ULONG_PTR)2;
-    pseudohandles[3] = GetCurrentProcessToken();
-    pseudohandles[4] = GetCurrentThreadToken();
-    pseudohandles[5] = GetCurrentThreadEffectiveToken();
 
     /* create the maximum number of events and make sure
      * we can wait on that many */
-    for (i = 0; i < MAXIMUM_WAIT_OBJECTS; i++)
+    for (i=0; i<MAXIMUM_WAIT_OBJECTS; i++)
     {
         maxevents[i] = CreateEventW(NULL, i==0, TRUE, NULL);
         ok( maxevents[i] != 0, "should create enough events\n");
@@ -1455,7 +1717,7 @@
     r = WaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, FALSE, 0);
     ok( r == WAIT_OBJECT_0, "should signal handle #0 first, got %ld\n", r);
     ok(ResetEvent(maxevents[0]), "ResetEvent\n");
-    for (i = 1; i < MAXIMUM_WAIT_OBJECTS; i++)
+    for (i=1; i<MAXIMUM_WAIT_OBJECTS; i++)
     {
         /* the lowest index is checked first and remaining events are untouched */
         r = WaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, FALSE, 0);
@@ -1463,96 +1725,50 @@
     }
 
     /* run same test with Nt* call */
-    for (i = 0; i < MAXIMUM_WAIT_OBJECTS; i++)
+    for (i=0; i<MAXIMUM_WAIT_OBJECTS; i++)
         SetEvent(maxevents[i]);
 
     /* a manual-reset event remains signaled, an auto-reset event is cleared */
-    status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, WaitAny, FALSE, NULL);
+    status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, TRUE, FALSE, NULL);
     ok(status == STATUS_WAIT_0, "should signal lowest handle first, got %08lx\n", status);
-    status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, WaitAny, FALSE, NULL);
+    status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, TRUE, FALSE, NULL);
     ok(status == STATUS_WAIT_0, "should signal handle #0 first, got %08lx\n", status);
     ok(ResetEvent(maxevents[0]), "ResetEvent\n");
-    for (i = 1; i < MAXIMUM_WAIT_OBJECTS; i++)
+    for (i=1; i<MAXIMUM_WAIT_OBJECTS; i++)
     {
         /* the lowest index is checked first and remaining events are untouched */
-        status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, WaitAny, FALSE, NULL);
+        status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, TRUE, FALSE, NULL);
         ok(status == STATUS_WAIT_0 + i, "should signal handle #%d first, got %08lx\n", i, status);
     }
 
-    CloseHandle(maxevents[0]);
-
-    /* in contrast to WaitForSingleObject, all pseudo-handles are not allowed in
+    for (i=0; i<MAXIMUM_WAIT_OBJECTS; i++)
+        if (maxevents[i]) CloseHandle(maxevents[i]);
+
+    /* in contrast to WaitForSingleObject, pseudo handles are not allowed in
      * WaitForMultipleObjects and NtWaitForMultipleObjects */
-    for (i = 0; i < ARRAY_SIZE(pseudohandles); i++)
-    {
-        maxevents[0] = pseudohandles[i];
-        SetLastError(0xdeadbeef);
-        r = WaitForMultipleObjects(1, maxevents, FALSE, 0);
-        ok(r == WAIT_FAILED, "expected WAIT_FAILED, got %lu\n", r);
-        ok(GetLastError() == ERROR_INVALID_HANDLE,
-           "expected ERROR_INVALID_HANDLE, got %lu\n", GetLastError());
-
-        status = pNtWaitForMultipleObjects(1, maxevents, WaitAny, FALSE, NULL);
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-
-        /* also not allowed with a wait count greater than 1 with both WaitAny and WaitAll */
-        SetLastError(0xdeadbeef);
-        r = WaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, FALSE, 0);
-        ok(r == WAIT_FAILED, "expected WAIT_FAILED, got %lu\n", r);
-        ok(GetLastError() == ERROR_INVALID_HANDLE,
-           "expected ERROR_INVALID_HANDLE, got %lu\n", GetLastError());
-        SetLastError(0xdeadbeef);
-        r = WaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, TRUE, 0);
-        ok(r == WAIT_FAILED, "expected WAIT_FAILED, got %lu\n", r);
-        ok(GetLastError() == ERROR_INVALID_HANDLE,
-           "expected ERROR_INVALID_HANDLE, got %lu\n", GetLastError());
-
-        status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, WaitAny, FALSE, NULL);
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-        status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, WaitAll, FALSE, NULL);
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-
-        /* irrespective of the index that contains the pseudo handle for both wait types */
-        maxevents[0] = maxevents[MAXIMUM_WAIT_OBJECTS - 1];
-        maxevents[MAXIMUM_WAIT_OBJECTS - 1] = pseudohandles[i];
-        SetLastError(0xdeadbeef);
-        r = WaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, FALSE, 0);
-        ok(r == WAIT_FAILED, "expected WAIT_FAILED, got %lu\n", r);
-        ok(GetLastError() == ERROR_INVALID_HANDLE,
-           "expected ERROR_INVALID_HANDLE, got %lu\n", GetLastError());
-        SetLastError(0xdeadbeef);
-        r = WaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, TRUE, 0);
-        ok(r == WAIT_FAILED, "expected WAIT_FAILED, got %lu\n", r);
-        ok(GetLastError() == ERROR_INVALID_HANDLE,
-           "expected ERROR_INVALID_HANDLE, got %lu\n", GetLastError());
-
-        status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, WaitAny, FALSE, NULL);
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-        status = pNtWaitForMultipleObjects(MAXIMUM_WAIT_OBJECTS, maxevents, WaitAll, FALSE, NULL);
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-        maxevents[MAXIMUM_WAIT_OBJECTS - 1] = maxevents[0];
-    }
-
-    /* Similar to WaitForSingleObject std handles are only allowed in WaitForMultipleObjects and
-     * not NtWaitForMultipleObjects (for brevity we only test single count waits here) */
-    for (i = 0; i < ARRAY_SIZE(std_handles); i++)
-    {
-        maxevents[0] = std_handles[i];
-        SetLastError(0xdeadbeef);
-        r = WaitForMultipleObjects(1, maxevents, FALSE, 0);
-        ok(r == WAIT_OBJECT_0 || r == WAIT_TIMEOUT,
-           "expected WAIT_OBJECT_0 or WAIT_TIMEOUT), got %lu, GetLastError()=%lu\n",
-           r, GetLastError());
-
-        status = pNtWaitForMultipleObjects(1, maxevents, WaitAny, FALSE, NULL);
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-
-        status = pNtWaitForMultipleObjects(1, maxevents, WaitAll, FALSE, NULL);
-        ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
-    }
-
-    for (i = 1; i < MAXIMUM_WAIT_OBJECTS; i++)
-        if (maxevents[i]) CloseHandle(maxevents[i]);
+    maxevents[0] = GetCurrentProcess();
+    SetLastError(0xdeadbeef);
+    r = WaitForMultipleObjects(1, maxevents, FALSE, 100);
+    todo_wine ok(r == WAIT_FAILED, "expected WAIT_FAILED, got %lu\n", r);
+    todo_wine ok(GetLastError() == ERROR_INVALID_HANDLE,
+                 "expected ERROR_INVALID_HANDLE, got %lu\n", GetLastError());
+
+    maxevents[0] = GetCurrentThread();
+    SetLastError(0xdeadbeef);
+    r = WaitForMultipleObjects(1, maxevents, FALSE, 100);
+    todo_wine ok(r == WAIT_FAILED, "expected WAIT_FAILED, got %lu\n", r);
+    todo_wine ok(GetLastError() == ERROR_INVALID_HANDLE,
+                 "expected ERROR_INVALID_HANDLE, got %lu\n", GetLastError());
+
+    timeout.QuadPart = -1000000;
+    maxevents[0] = GetCurrentProcess();
+    status = pNtWaitForMultipleObjects(1, maxevents, TRUE, FALSE, &timeout);
+    todo_wine ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
+
+    timeout.QuadPart = -1000000;
+    maxevents[0] = GetCurrentThread();
+    status = pNtWaitForMultipleObjects(1, maxevents, TRUE, FALSE, &timeout);
+    todo_wine ok(status == STATUS_INVALID_HANDLE, "expected STATUS_INVALID_HANDLE, got %08lx\n", status);
 }
 
 static BOOL g_initcallback_ret, g_initcallback_called;
@@ -1936,13 +2152,13 @@
 static CRITICAL_SECTION condvar_crit;
 static SRWLOCK condvar_srwlock;
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 static struct
 {
     char c;
     CONDITION_VARIABLE cv;
 } unaligned_cv;
-#pragma pack(pop)
+#include "poppack.h"
 
 /* Sequence of wake/sleep to check boundary conditions:
  * 0: init
@@ -2172,13 +2388,13 @@
 } srwlock_base_errors;
 
 #if defined(__i386__) || defined(__x86_64__)
-#pragma pack(push,1)
+#include "pshpack1.h"
 struct
 {
     char c;
     SRWLOCK lock;
 } unaligned_srwlock;
-#pragma pack(pop)
+#include "poppack.h"
 #endif
 
 /* Sequence of acquire/release to check boundary conditions:
@@ -2731,15 +2947,15 @@
 
     ReleaseSemaphore(semaphores[0], 1, NULL);
     timeout.QuadPart = -10000000;
-    status = pNtWaitForMultipleObjects(1, &semaphores[1], WaitAll, TRUE, &timeout);
+    status = pNtWaitForMultipleObjects(1, &semaphores[1], FALSE, TRUE, &timeout);
     ok(status == STATUS_USER_APC, "expected STATUS_USER_APC, got %08lx\n", status);
     timeout.QuadPart = -2000000;
-    status = pNtWaitForMultipleObjects(1, &semaphores[1], WaitAll, TRUE, &timeout);
+    status = pNtWaitForMultipleObjects(1, &semaphores[1], FALSE, TRUE, &timeout);
     ok(status == STATUS_WAIT_0, "expected STATUS_WAIT_0, got %08lx\n", status);
 
     ReleaseSemaphore(semaphores[0], 1, NULL);
     timeout.QuadPart = -10000000;
-    status = pNtWaitForMultipleObjects(1, &semaphores[1], WaitAll, TRUE, &timeout);
+    status = pNtWaitForMultipleObjects(1, &semaphores[1], FALSE, TRUE, &timeout);
     ok(status == STATUS_USER_APC, "expected STATUS_USER_APC, got %08lx\n", status);
     result = WaitForSingleObject(semaphores[0], 0);
     ok(result == WAIT_TIMEOUT, "expected WAIT_TIMEOUT, got %lu\n", result);
@@ -3040,8 +3256,6 @@
     ok(ret == STATUS_UNSUCCESSFUL, "got %#lx\n", ret);
     ret = pNtQueueApcThread(thread, NULL, 0, 0, 0);
     ok(ret == STATUS_UNSUCCESSFUL, "got %#lx\n", ret);
-    ret = pNtQueueApcThread((HANDLE)0xdeadbeef, call_user_apc, (ULONG_PTR)user_apc, 0, 0);
-    ok(ret == STATUS_INVALID_HANDLE, "got %#lx\n", ret);
 
     SetLastError(0xdeadbeef);
     ret = QueueUserAPC(user_apc, thread, 0);
@@ -3073,33 +3287,84 @@
     status = pNtTestAlert();
     ok(!status, "got %lx\n", status);
     ok(apc_count == 1, "APC count %u\n", apc_count);
-
-    if (!pQueueUserAPC2)
-    {
-        win_skip("QueueUserAPC2 is not available.\n");
-        return;
-    }
-
-    apc_count = 0;
-    ret = pQueueUserAPC2(user_apc, GetCurrentThread(), 0, QUEUE_USER_APC_FLAGS_NONE);
-    ok(ret, "QueueUserAPC failed err %lu\n", GetLastError());
-    ok(!apc_count, "got %d.\n", apc_count);
-    SleepEx( 0, TRUE );
-    ok(apc_count == 1, "got %d.\n", apc_count);
-
-    apc_count = 0;
-    ret = pQueueUserAPC2(user_apc, GetCurrentThread(), 0, QUEUE_USER_APC_FLAGS_SPECIAL_USER_APC);
-    if (sizeof(void *) == 4)
-        ok(!ret && GetLastError() == ERROR_NOT_SUPPORTED, "got ret %lu, error %lu.\n", ret, GetLastError());
-    else
-        ok(ret, "got error %lu.\n", GetLastError());
-
-    if (ret)
-    {
-        todo_wine ok(apc_count == 1, "got %d.\n", apc_count);
-        SleepEx( 0, TRUE );
-        ok(apc_count == 1, "got %d.\n", apc_count);
-    }
+}
+
+static int zigzag_state, zigzag_count[2], zigzag_stop;
+
+static DWORD CALLBACK zigzag_event0(void *arg)
+{
+    HANDLE *events = arg;
+
+    while (!zigzag_stop)
+    {
+        WaitForSingleObject(events[0], INFINITE);
+        ResetEvent(events[0]);
+        ok(zigzag_state == 0, "got wrong state %d\n", zigzag_state);
+        zigzag_state++;
+        SetEvent(events[1]);
+        zigzag_count[0]++;
+    }
+    trace("thread 0 got done\n");
+    return 0;
+}
+
+static DWORD CALLBACK zigzag_event1(void *arg)
+{
+    HANDLE *events = arg;
+
+    while (!zigzag_stop)
+    {
+        WaitForSingleObject(events[1], INFINITE);
+        ResetEvent(events[1]);
+        ok(zigzag_state == 1, "got wrong state %d\n", zigzag_state);
+        zigzag_state--;
+        SetEvent(events[0]);
+        zigzag_count[1]++;
+    }
+    trace("thread 1 got done\n");
+    return 0;
+}
+
+static void test_zigzag_event(void)
+{
+    /* The basic idea is to test SetEvent/Wait back and forth between two
+     * threads. Each thread clears their own event, sets some common data,
+     * signals the other's, then waits on their own. We make sure the common
+     * data is always in the right state. We also print performance data. */
+
+    HANDLE threads[2], events[2];
+    BOOL ret;
+
+    events[0] = CreateEventA(NULL, FALSE, FALSE, NULL);
+    events[1] = CreateEventA(NULL, FALSE, FALSE, NULL);
+
+    threads[0] = CreateThread(NULL, 0, zigzag_event0, events, 0, NULL);
+    threads[1] = CreateThread(NULL, 0, zigzag_event1, events, 0, NULL);
+
+    zigzag_state = 0;
+    zigzag_count[0] = zigzag_count[1] = 0;
+    zigzag_stop = 0;
+
+    trace("starting zigzag test (events)\n");
+    SetEvent(events[0]);
+    Sleep(2000);
+    zigzag_stop = 1;
+    ret = WaitForMultipleObjects(2, threads, FALSE, INFINITE);
+    trace("%d\n", ret);
+    ok(ret == 0 || ret == 1, "wait failed: %u\n", ret);
+
+    ok(zigzag_count[0] == zigzag_count[1] || zigzag_count[0] == zigzag_count[1] + 1,
+        "count did not match: %d != %d\n", zigzag_count[0], zigzag_count[1]);
+
+    /* signal the other thread to finish, if it didn't already
+     * (in theory they both would at the same time, but there's a slight race on teardown if we get
+     * thread 1 SetEvent -> thread 0 ResetEvent -> thread 0 Wait -> thread 1 exits */
+    zigzag_state = 1-ret;
+    SetEvent(events[1-ret]);
+    ret = WaitForSingleObject(threads[1-ret], 1000);
+    ok(!ret, "wait failed: %u\n", ret);
+
+    trace("count: %d\n", zigzag_count[0]);
 }
 
 struct test_barrier_thread_param
@@ -3385,12 +3650,12 @@
     pReleaseSRWLockShared = (void *)GetProcAddress(hdll, "ReleaseSRWLockShared");
     pTryAcquireSRWLockExclusive = (void *)GetProcAddress(hdll, "TryAcquireSRWLockExclusive");
     pTryAcquireSRWLockShared = (void *)GetProcAddress(hdll, "TryAcquireSRWLockShared");
-    pQueueUserAPC2 = (void *)GetProcAddress(hdll, "QueueUserAPC2");
     pInitializeSynchronizationBarrier = (void *)GetProcAddress(hdll, "InitializeSynchronizationBarrier");
     pDeleteSynchronizationBarrier = (void *)GetProcAddress(hdll, "DeleteSynchronizationBarrier");
     pEnterSynchronizationBarrier = (void *)GetProcAddress(hdll, "EnterSynchronizationBarrier");
     pNtAllocateVirtualMemory = (void *)GetProcAddress(hntdll, "NtAllocateVirtualMemory");
     pNtFreeVirtualMemory = (void *)GetProcAddress(hntdll, "NtFreeVirtualMemory");
+    pNtQuerySystemTime = (void *)GetProcAddress(hntdll, "NtQuerySystemTime");
     pNtWaitForSingleObject = (void *)GetProcAddress(hntdll, "NtWaitForSingleObject");
     pNtWaitForMultipleObjects = (void *)GetProcAddress(hntdll, "NtWaitForMultipleObjects");
     pRtlInterlockedPushListSList = (void *)GetProcAddress(hntdll, "RtlInterlockedPushListSList");
@@ -3435,6 +3700,7 @@
     test_srwlock_example();
     test_alertable_wait();
     test_apc_deadlock();
+    test_zigzag_event();
     test_crit_section();
     test_barrier();
 }
