diff --git a/dlls/amstream/ddrawstream.c b/dlls/amstream/ddrawstream.c
--- a/dlls/amstream/ddrawstream.c
+++ b/dlls/amstream/ddrawstream.c
@@ -41,7 +41,6 @@
     IAMMediaStream IAMMediaStream_iface;
     IDirectDrawMediaStream IDirectDrawMediaStream_iface;
     IMemInputPin IMemInputPin_iface;
-    IMemAllocator IMemAllocator_iface;
     IPin IPin_iface;
     LONG ref;
     LONG sample_refs;
@@ -55,7 +54,7 @@
     IFilterGraph *graph;
 
     IPin *peer;
-    BOOL using_private_allocator;
+    IMemAllocator *allocator;
     AM_MEDIA_TYPE mt;
     struct format format;
     FILTER_STATE state;
@@ -64,10 +63,6 @@
     BOOL flushing;
     CONDITION_VARIABLE update_queued_cv;
     struct list update_queue;
-
-    CONDITION_VARIABLE allocator_cv;
-    bool committed;
-    LONG buffer_count; /* Only used for properties. */
 };
 
 struct ddraw_sample
@@ -77,7 +72,6 @@
     struct ddraw_stream *parent;
     IMultiMediaStream *mmstream;
     IDirectDrawSurface *surface;
-    DDSURFACEDESC surface_desc;
     RECT rect;
     STREAM_TIME start_time;
     STREAM_TIME end_time;
@@ -85,13 +79,9 @@
     CONDITION_VARIABLE update_cv;
     HANDLE external_event;
 
-    IMediaSample IMediaSample_iface;
-    unsigned int media_sample_refcount;
-    bool sync_point, preroll, discontinuity;
-
     struct list entry;
     HRESULT update_hr;
-    bool pending;
+    BOOL busy;
 };
 
 static HRESULT ddrawstreamsample_create(struct ddraw_stream *parent, IDirectDrawSurface *surface,
@@ -99,7 +89,7 @@
 
 static void remove_queued_update(struct ddraw_sample *sample)
 {
-    sample->pending = false;
+    sample->busy = FALSE;
     list_remove(&sample->entry);
     WakeConditionVariable(&sample->update_cv);
     if (sample->external_event)
@@ -117,7 +107,7 @@
     }
 }
 
-static HRESULT copy_sample(struct ddraw_sample *sample, int stride, BYTE *pointer,
+static HRESULT process_update(struct ddraw_sample *sample, int stride, BYTE *pointer,
         STREAM_TIME start_time, STREAM_TIME end_time)
 {
     DDSURFACEDESC desc;
@@ -209,12 +199,6 @@
     {
         IAMMediaStream_AddRef(iface);
         *ret_iface = &This->IMemInputPin_iface;
-        return S_OK;
-    }
-    else if (IsEqualGUID(riid, &IID_IMemAllocator))
-    {
-        IAMMediaStream_AddRef(iface);
-        *ret_iface = &This->IMemAllocator_iface;
         return S_OK;
     }
 
@@ -548,73 +532,11 @@
     return S_OK;
 }
 
-static unsigned int align(unsigned int n, unsigned int alignment)
-{
-    return (n + alignment - 1) & ~(alignment - 1);
-}
-
-static void set_mt_from_desc(AM_MEDIA_TYPE *mt, const DDSURFACEDESC *format, unsigned int pitch)
-{
-    VIDEOINFO *videoinfo = CoTaskMemAlloc(sizeof(VIDEOINFO));
-
-    memset(mt, 0, sizeof(*mt));
-    mt->majortype = MEDIATYPE_Video;
-    mt->formattype = FORMAT_VideoInfo;
-    mt->cbFormat = sizeof(VIDEOINFO);
-    mt->pbFormat = (BYTE *)videoinfo;
-
-    memset(videoinfo, 0, sizeof(*videoinfo));
-    SetRect(&videoinfo->rcSource, 0, 0, format->dwWidth, format->dwHeight);
-    SetRect(&videoinfo->rcTarget, 0, 0, format->dwWidth, format->dwHeight);
-    videoinfo->bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
-    videoinfo->bmiHeader.biWidth = pitch * 8 / format->ddpfPixelFormat.dwRGBBitCount;
-    videoinfo->bmiHeader.biHeight = -format->dwHeight;
-    videoinfo->bmiHeader.biBitCount = format->ddpfPixelFormat.dwRGBBitCount;
-    videoinfo->bmiHeader.biCompression = BI_RGB;
-    videoinfo->bmiHeader.biPlanes = 1;
-    videoinfo->bmiHeader.biSizeImage = align(pitch * format->dwHeight, 4);
-
-    mt->lSampleSize = videoinfo->bmiHeader.biSizeImage;
-    mt->bFixedSizeSamples = TRUE;
-
-    if (format->ddpfPixelFormat.dwRGBBitCount == 16 && format->ddpfPixelFormat.dwRBitMask == 0x7c00)
-    {
-        mt->subtype = MEDIASUBTYPE_RGB555;
-    }
-    else if (format->ddpfPixelFormat.dwRGBBitCount == 16 && format->ddpfPixelFormat.dwRBitMask == 0xf800)
-    {
-        mt->subtype = MEDIASUBTYPE_RGB565;
-        videoinfo = (VIDEOINFO *)mt->pbFormat;
-        videoinfo->bmiHeader.biCompression = BI_BITFIELDS;
-        videoinfo->dwBitMasks[iRED]   = 0xf800;
-        videoinfo->dwBitMasks[iGREEN] = 0x07e0;
-        videoinfo->dwBitMasks[iBLUE]  = 0x001f;
-    }
-    else if (format->ddpfPixelFormat.dwRGBBitCount == 24)
-    {
-        mt->subtype = MEDIASUBTYPE_RGB24;
-    }
-    else if (format->ddpfPixelFormat.dwRGBBitCount == 32)
-    {
-        mt->subtype = MEDIASUBTYPE_RGB32;
-    }
-    else if (format->ddpfPixelFormat.dwRGBBitCount == 8 && (format->ddpfPixelFormat.dwFlags & DDPF_PALETTEINDEXED8))
-    {
-        mt->subtype = MEDIASUBTYPE_RGB8;
-        videoinfo->bmiHeader.biClrUsed = 256;
-        /* FIXME: Translate the palette. */
-    }
-    else
-    {
-        FIXME("Unknown flags %#lx, bit count %lu.\n",
-                format->ddpfPixelFormat.dwFlags, format->ddpfPixelFormat.dwRGBBitCount);
-    }
-}
-
 static HRESULT WINAPI ddraw_IDirectDrawMediaStream_SetFormat(IDirectDrawMediaStream *iface,
         const DDSURFACEDESC *format, IDirectDrawPalette *palette)
 {
     struct ddraw_stream *stream = impl_from_IDirectDrawMediaStream(iface);
+    AM_MEDIA_TYPE old_media_type;
     struct format old_format;
     IPin *old_peer;
     HRESULT hr;
@@ -637,83 +559,49 @@
     if (format->dwFlags & DDSD_PIXELFORMAT)
     {
         if (format->ddpfPixelFormat.dwSize != sizeof(DDPIXELFORMAT))
-        {
-            WARN("Invalid size %#lx, returning DDERR_INVALIDSURFACETYPE.\n", format->ddpfPixelFormat.dwSize);
             return DDERR_INVALIDSURFACETYPE;
-        }
 
         if (format->ddpfPixelFormat.dwFlags & DDPF_FOURCC)
         {
             if (!format->ddpfPixelFormat.dwRGBBitCount)
-            {
-                WARN("Invalid zero bit count, returning E_INVALIDARG.\n");
                 return E_INVALIDARG;
-            }
         }
         else
         {
             if (format->ddpfPixelFormat.dwFlags & (DDPF_YUV | DDPF_PALETTEINDEXED1 |
                     DDPF_PALETTEINDEXED2 | DDPF_PALETTEINDEXED4 | DDPF_PALETTEINDEXEDTO8))
-            {
-                WARN("Rejecting flags %#lx.\n", format->ddpfPixelFormat.dwFlags);
                 return DDERR_INVALIDSURFACETYPE;
-            }
 
             if (!(format->ddpfPixelFormat.dwFlags & DDPF_RGB))
-            {
-                WARN("Rejecting non-RGB flags %#lx.\n", format->ddpfPixelFormat.dwFlags);
                 return DDERR_INVALIDSURFACETYPE;
-            }
 
             switch (format->ddpfPixelFormat.dwRGBBitCount)
             {
             case 8:
                 if (!(format->ddpfPixelFormat.dwFlags & DDPF_PALETTEINDEXED8))
-                {
-                    WARN("Rejecting non-palettized 8-bit format.\n");
                     return DDERR_INVALIDSURFACETYPE;
-                }
                 break;
             case 16:
                 if (format->ddpfPixelFormat.dwFlags & DDPF_PALETTEINDEXED8)
-                {
-                    WARN("Rejecting palettized 16-bit format.\n");
                     return DDERR_INVALIDSURFACETYPE;
-                }
                 if ((format->ddpfPixelFormat.dwRBitMask != 0x7c00 ||
                     format->ddpfPixelFormat.dwGBitMask != 0x03e0 ||
                     format->ddpfPixelFormat.dwBBitMask != 0x001f) &&
                     (format->ddpfPixelFormat.dwRBitMask != 0xf800 ||
                     format->ddpfPixelFormat.dwGBitMask != 0x07e0 ||
                     format->ddpfPixelFormat.dwBBitMask != 0x001f))
-                {
-                    WARN("Rejecting bit masks %08lx, %08lx, %08lx.\n",
-                            format->ddpfPixelFormat.dwRBitMask,
-                            format->ddpfPixelFormat.dwGBitMask,
-                            format->ddpfPixelFormat.dwBBitMask);
                     return DDERR_INVALIDSURFACETYPE;
-                }
                 break;
             case 24:
             case 32:
                 if (format->ddpfPixelFormat.dwFlags & DDPF_PALETTEINDEXED8)
-                {
-                    WARN("Rejecting palettized %lu-bit format.\n", format->ddpfPixelFormat.dwRGBBitCount);
                     return DDERR_INVALIDSURFACETYPE;
-                }
                 if (format->ddpfPixelFormat.dwRBitMask != 0xff0000 ||
                     format->ddpfPixelFormat.dwGBitMask != 0x00ff00 ||
                     format->ddpfPixelFormat.dwBBitMask != 0x0000ff)
-                {
-                    WARN("Rejecting bit masks %08lx, %08lx, %08lx.\n",
-                            format->ddpfPixelFormat.dwRBitMask,
-                            format->ddpfPixelFormat.dwGBitMask,
-                            format->ddpfPixelFormat.dwBBitMask);
                     return DDERR_INVALIDSURFACETYPE;
-                }
                 break;
             default:
-                WARN("Unknown bit count %lu.\n", format->ddpfPixelFormat.dwRGBBitCount);
                 return DDERR_INVALIDSURFACETYPE;
             }
         }
@@ -733,44 +621,31 @@
 
     if (stream->peer && !is_format_compatible(stream, old_format.width, old_format.height, &old_format.pf))
     {
-        AM_MEDIA_TYPE new_mt;
-
-        if (stream->sample_refs > 0)
+        hr = CopyMediaType(&old_media_type, &stream->mt);
+        if (FAILED(hr))
         {
-            WARN("Outstanding sample references, returning MS_E_SAMPLEALLOC.\n");
             stream->format = old_format;
             LeaveCriticalSection(&stream->cs);
-            return MS_E_SAMPLEALLOC;
+            return hr;
         }
-
-        set_mt_from_desc(&new_mt, format, format->dwWidth * format->ddpfPixelFormat.dwRGBBitCount / 8);
-
-        if (!stream->using_private_allocator || IPin_QueryAccept(stream->peer, &new_mt) != S_OK)
+        old_peer = stream->peer;
+        IPin_AddRef(old_peer);
+
+        IFilterGraph_Disconnect(stream->graph, stream->peer);
+        IFilterGraph_Disconnect(stream->graph, &stream->IPin_iface);
+        hr = IFilterGraph_ConnectDirect(stream->graph, old_peer, &stream->IPin_iface, NULL);
+        if (FAILED(hr))
         {
-            AM_MEDIA_TYPE old_mt;
-
-            /* Reconnect. */
-            old_peer = stream->peer;
-            IPin_AddRef(old_peer);
-            CopyMediaType(&old_mt, &stream->mt);
-
-            IFilterGraph_Disconnect(stream->graph, stream->peer);
-            IFilterGraph_Disconnect(stream->graph, &stream->IPin_iface);
-            if (FAILED(hr = IFilterGraph_ConnectDirect(stream->graph, old_peer, &stream->IPin_iface, NULL)))
-            {
-                stream->format = old_format;
-                IFilterGraph_ConnectDirect(stream->graph, old_peer, &stream->IPin_iface, &old_mt);
-                IPin_Release(old_peer);
-                FreeMediaType(&old_mt);
-                FreeMediaType(&new_mt);
-                LeaveCriticalSection(&stream->cs);
-                return DDERR_INVALIDSURFACETYPE;
-            }
-            FreeMediaType(&old_mt);
+            stream->format = old_format;
+            IFilterGraph_ConnectDirect(stream->graph, old_peer, &stream->IPin_iface, &old_media_type);
             IPin_Release(old_peer);
+            FreeMediaType(&old_media_type);
+            LeaveCriticalSection(&stream->cs);
+            return DDERR_INVALIDSURFACETYPE;
         }
 
-        FreeMediaType(&new_mt);
+        IPin_Release(old_peer);
+        FreeMediaType(&old_media_type);
     }
 
     LeaveCriticalSection(&stream->cs);
@@ -1062,7 +937,6 @@
     DDPIXELFORMAT pf = {sizeof(DDPIXELFORMAT)};
 
     TRACE("stream %p, peer %p, mt %p.\n", stream, peer, mt);
-    strmbase_dump_media_type(mt);
 
     EnterCriticalSection(&stream->cs);
 
@@ -1400,186 +1274,6 @@
     ddraw_sink_NewSegment,
 };
 
-static struct ddraw_stream *impl_from_IMemAllocator(IMemAllocator *iface)
-{
-    return CONTAINING_RECORD(iface, struct ddraw_stream, IMemAllocator_iface);
-}
-
-static HRESULT WINAPI ddraw_mem_allocator_QueryInterface(IMemAllocator *iface, REFIID iid, void **out)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-    return IAMMediaStream_QueryInterface(&stream->IAMMediaStream_iface, iid, out);
-}
-
-static ULONG WINAPI ddraw_mem_allocator_AddRef(IMemAllocator *iface)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-    return IAMMediaStream_AddRef(&stream->IAMMediaStream_iface);
-}
-
-static ULONG WINAPI ddraw_mem_allocator_Release(IMemAllocator *iface)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-    return IAMMediaStream_Release(&stream->IAMMediaStream_iface);
-}
-
-static HRESULT WINAPI ddraw_mem_allocator_SetProperties(IMemAllocator *iface,
-        ALLOCATOR_PROPERTIES *req_props, ALLOCATOR_PROPERTIES *ret_props)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-
-    TRACE("stream %p, req_props %p, ret_props %p.\n", stream, req_props, ret_props);
-
-    TRACE("Requested %ld buffers, size %ld, prefix %ld, alignment %ld.\n",
-            req_props->cBuffers, req_props->cbBuffer, req_props->cbPrefix, req_props->cbAlign);
-
-    if (!req_props->cbAlign)
-        return VFW_E_BADALIGN;
-
-    EnterCriticalSection(&stream->cs);
-
-    if (stream->committed)
-    {
-        LeaveCriticalSection(&stream->cs);
-        return VFW_E_ALREADY_COMMITTED;
-    }
-
-    stream->buffer_count = max(req_props->cBuffers, 1);
-
-    ret_props->cBuffers = stream->buffer_count;
-    ret_props->cbBuffer = stream->format.width * stream->format.height * stream->format.pf.dwRGBBitCount / 8;
-    ret_props->cbAlign = 1;
-    ret_props->cbPrefix = 0;
-
-    LeaveCriticalSection(&stream->cs);
-
-    return S_OK;
-}
-
-static HRESULT WINAPI ddraw_mem_allocator_GetProperties(IMemAllocator *iface, ALLOCATOR_PROPERTIES *props)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-
-    TRACE("stream %p, props %p.\n", stream, props);
-
-    EnterCriticalSection(&stream->cs);
-    props->cBuffers = stream->buffer_count;
-    props->cbBuffer = stream->format.width * stream->format.height * stream->format.pf.dwRGBBitCount / 8;
-    props->cbAlign = 1;
-    props->cbPrefix = 0;
-    LeaveCriticalSection(&stream->cs);
-
-    return S_OK;
-}
-
-static HRESULT WINAPI ddraw_mem_allocator_Commit(IMemAllocator *iface)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-
-    TRACE("stream %p.\n", stream);
-
-    EnterCriticalSection(&stream->cs);
-    stream->committed = true;
-    /* We have nothing to actually commit; all of our samples are created by
-     * CreateSample(). */
-    LeaveCriticalSection(&stream->cs);
-
-    return S_OK;
-}
-
-static HRESULT WINAPI ddraw_mem_allocator_Decommit(IMemAllocator *iface)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-
-    TRACE("stream %p.\n", stream);
-
-    EnterCriticalSection(&stream->cs);
-    stream->committed = false;
-    WakeAllConditionVariable(&stream->allocator_cv);
-    /* We have nothing to actually decommit; all of our samples are created by
-     * CreateSample(). */
-    LeaveCriticalSection(&stream->cs);
-
-    return S_OK;
-}
-
-static struct ddraw_sample *get_pending_sample(struct ddraw_stream *stream)
-{
-    struct ddraw_sample *sample;
-
-    LIST_FOR_EACH_ENTRY(sample, &stream->update_queue, struct ddraw_sample, entry)
-    {
-        if (!sample->media_sample_refcount)
-            return sample;
-    }
-
-    return NULL;
-}
-
-static HRESULT WINAPI ddraw_mem_allocator_GetBuffer(IMemAllocator *iface,
-        IMediaSample **ret_sample, REFERENCE_TIME *start, REFERENCE_TIME *end, DWORD flags)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-    struct ddraw_sample *sample;
-    HRESULT hr;
-
-    TRACE("stream %p, ret_sample %p, start %p, end %p, flags %#lx.\n", stream, ret_sample, start, end, flags);
-
-    EnterCriticalSection(&stream->cs);
-
-    while (stream->committed && !(sample = get_pending_sample(stream)))
-        SleepConditionVariableCS(&stream->allocator_cv, &stream->cs, INFINITE);
-
-    if (!stream->committed)
-    {
-        LeaveCriticalSection(&stream->cs);
-        return VFW_E_NOT_COMMITTED;
-    }
-
-    sample->surface_desc.dwSize = sizeof(DDSURFACEDESC);
-    /* Don't pass the sample rect here; the upstream filter is expected to
-     * deal with it. */
-    if ((FAILED(hr = IDirectDrawSurface_Lock(sample->surface,
-            NULL, &sample->surface_desc, DDLOCK_WAIT, NULL))))
-    {
-        LeaveCriticalSection(&stream->cs);
-        return hr;
-    }
-
-    /* Only these fields are reset. */
-    sample->sync_point = true;
-    sample->discontinuity = false;
-
-    sample->media_sample_refcount = 1;
-
-    LeaveCriticalSection(&stream->cs);
-
-    *ret_sample = &sample->IMediaSample_iface;
-    return S_OK;
-}
-
-static HRESULT WINAPI ddraw_mem_allocator_ReleaseBuffer(IMemAllocator *iface, IMediaSample *sample)
-{
-    struct ddraw_stream *stream = impl_from_IMemAllocator(iface);
-
-    FIXME("stream %p, sample %p, stub!\n", stream, sample);
-
-    return E_NOTIMPL;
-}
-
-static const IMemAllocatorVtbl ddraw_mem_allocator_vtbl =
-{
-    ddraw_mem_allocator_QueryInterface,
-    ddraw_mem_allocator_AddRef,
-    ddraw_mem_allocator_Release,
-    ddraw_mem_allocator_SetProperties,
-    ddraw_mem_allocator_GetProperties,
-    ddraw_mem_allocator_Commit,
-    ddraw_mem_allocator_Decommit,
-    ddraw_mem_allocator_GetBuffer,
-    ddraw_mem_allocator_ReleaseBuffer,
-};
-
 static inline struct ddraw_stream *impl_from_IMemInputPin(IMemInputPin *iface)
 {
     return CONTAINING_RECORD(iface, struct ddraw_stream, IMemInputPin_iface);
@@ -1609,8 +1303,14 @@
 
     TRACE("stream %p, allocator %p.\n", stream, allocator);
 
-    IMemAllocator_AddRef(*allocator = &stream->IMemAllocator_iface);
-    return S_OK;
+    if (stream->allocator)
+    {
+        IMemAllocator_AddRef(*allocator = stream->allocator);
+        return S_OK;
+    }
+
+    *allocator = NULL;
+    return VFW_E_NO_ALLOCATOR;
 }
 
 static HRESULT WINAPI ddraw_meminput_NotifyAllocator(IMemInputPin *iface, IMemAllocator *allocator, BOOL readonly)
@@ -1622,7 +1322,11 @@
     if (!allocator)
         return E_POINTER;
 
-    stream->using_private_allocator = (allocator == &stream->IMemAllocator_iface);
+    if (allocator)
+        IMemAllocator_AddRef(allocator);
+    if (stream->allocator)
+        IMemAllocator_Release(stream->allocator);
+    stream->allocator = allocator;
 
     return S_OK;
 }
@@ -1633,81 +1337,54 @@
     return E_NOTIMPL;
 }
 
-static struct ddraw_sample *get_update_sample(struct ddraw_stream *stream, IMediaSample *buffer)
-{
-    const BITMAPINFOHEADER *bitmap_info = &((VIDEOINFOHEADER *)stream->mt.pbFormat)->bmiHeader;
-    STREAM_TIME start_stream_time, end_stream_time;
-    REFERENCE_TIME start_time = 0, end_time = 0;
-    struct ddraw_sample *sample;
+static HRESULT WINAPI ddraw_meminput_Receive(IMemInputPin *iface, IMediaSample *sample)
+{
+    struct ddraw_stream *stream = impl_from_IMemInputPin(iface);
+    BITMAPINFOHEADER *bitmap_info;
+    REFERENCE_TIME start_time = 0;
+    REFERENCE_TIME end_time = 0;
+    STREAM_TIME start_stream_time;
+    STREAM_TIME end_stream_time;
+    IMediaStreamFilter *filter;
+    STREAM_TIME current_time;
     BYTE *top_down_pointer;
     int top_down_stride;
     BYTE *pointer;
+    BOOL top_down;
     int stride;
-
-    /* Is it any of the samples we gave out from GetBuffer()? */
-    LIST_FOR_EACH_ENTRY(sample, &stream->update_queue, struct ddraw_sample, entry)
-    {
-        if (buffer == &sample->IMediaSample_iface)
-        {
-            sample->update_hr = S_OK;
-            return sample;
-        }
-    }
-
-    /* Find an unused sample and blit to it. */
-
-    IMediaSample_GetPointer(buffer, &pointer);
-    IMediaSample_GetTime(buffer, &start_time, &end_time);
+    HRESULT hr;
+
+    TRACE("stream %p, sample %p.\n", stream, sample);
+
+    hr = IMediaSample_GetPointer(sample, &pointer);
+    if (FAILED(hr))
+        return hr;
+
+    IMediaSample_GetTime(sample, &start_time, &end_time);
+
+    EnterCriticalSection(&stream->cs);
+
+    if (stream->state == State_Stopped)
+    {
+        LeaveCriticalSection(&stream->cs);
+        return S_OK;
+    }
+    if (stream->flushing)
+    {
+        LeaveCriticalSection(&stream->cs);
+        return S_FALSE;
+    }
+
+    bitmap_info = &((VIDEOINFOHEADER *)stream->mt.pbFormat)->bmiHeader;
+
+    stride = ((bitmap_info->biWidth * bitmap_info->biBitCount + 31) & ~31) / 8;
+    top_down = (bitmap_info->biHeight < 0);
+
+    top_down_stride = top_down ? stride : -stride;
+    top_down_pointer = top_down ? pointer : pointer + stride * (bitmap_info->biHeight - 1);
 
     start_stream_time = start_time + stream->segment_start;
     end_stream_time = end_time + stream->segment_start;
-
-    stride = ((bitmap_info->biWidth * bitmap_info->biBitCount + 31) & ~31) / 8;
-    if (bitmap_info->biHeight < 0)
-    {
-        top_down_stride = stride;
-        top_down_pointer = pointer;
-    }
-    else
-    {
-        top_down_stride = -stride;
-        top_down_pointer = pointer + stride * (bitmap_info->biHeight - 1);
-    }
-
-    if ((sample = get_pending_sample(stream)))
-    {
-        sample->update_hr = copy_sample(sample, top_down_stride, top_down_pointer,
-                start_stream_time, end_stream_time);
-        return sample;
-    }
-
-    /* We don't have a sample yet. */
-    return NULL;
-}
-
-static HRESULT WINAPI ddraw_meminput_Receive(IMemInputPin *iface, IMediaSample *buffer)
-{
-    struct ddraw_stream *stream = impl_from_IMemInputPin(iface);
-    REFERENCE_TIME start_time = 0, end_time = 0;
-    IMediaStreamFilter *filter;
-    STREAM_TIME current_time;
-
-    TRACE("stream %p, buffer %p.\n", stream, buffer);
-
-    IMediaSample_GetTime(buffer, &start_time, &end_time);
-
-    EnterCriticalSection(&stream->cs);
-
-    if (stream->state == State_Stopped)
-    {
-        LeaveCriticalSection(&stream->cs);
-        return S_OK;
-    }
-    if (stream->flushing)
-    {
-        LeaveCriticalSection(&stream->cs);
-        return S_FALSE;
-    }
 
     filter = stream->filter;
 
@@ -1719,9 +1396,6 @@
 
     for (;;)
     {
-        struct ddraw_sample *sample;
-        IQualityControl *qc;
-
         if (stream->state == State_Stopped)
         {
             LeaveCriticalSection(&stream->cs);
@@ -1732,9 +1406,14 @@
             LeaveCriticalSection(&stream->cs);
             return S_FALSE;
         }
-
-        if ((sample = get_update_sample(stream, buffer)))
+        if (!list_empty(&stream->update_queue))
         {
+            struct ddraw_sample *sample = LIST_ENTRY(list_head(&stream->update_queue), struct ddraw_sample, entry);
+            IQualityControl *qc;
+
+            sample->update_hr = process_update(sample, top_down_stride, top_down_pointer,
+                    start_stream_time, end_stream_time);
+
             if (sample->continuous_update && SUCCEEDED(sample->update_hr))
             {
                 list_remove(&sample->entry);
@@ -1745,7 +1424,7 @@
                 remove_queued_update(sample);
             }
 
-            if (IMediaStreamFilter_GetCurrentStreamTime(filter, &current_time) == S_OK
+            if (S_OK == IMediaStreamFilter_GetCurrentStreamTime(filter, &current_time)
                     && SUCCEEDED(IPin_QueryInterface(stream->peer, &IID_IQualityControl, (void **)&qc)))
             {
                 Quality q;
@@ -1804,17 +1483,11 @@
     object->IAMMediaStream_iface.lpVtbl = &ddraw_IAMMediaStream_vtbl;
     object->IDirectDrawMediaStream_iface.lpVtbl = &ddraw_IDirectDrawMediaStream_Vtbl;
     object->IMemInputPin_iface.lpVtbl = &ddraw_meminput_vtbl;
-    object->IMemAllocator_iface.lpVtbl = &ddraw_mem_allocator_vtbl;
     object->IPin_iface.lpVtbl = &ddraw_sink_vtbl;
-    object->IMemAllocator_iface.lpVtbl = &ddraw_mem_allocator_vtbl;
     object->ref = 1;
 
     object->format.width = 100;
     object->format.height = 100;
-
-    object->using_private_allocator = TRUE;
-    InitializeConditionVariable(&object->allocator_cv);
-    object->buffer_count = 1;
 
     InitializeCriticalSection(&object->cs);
     InitializeConditionVariable(&object->update_queued_cv);
@@ -1873,13 +1546,6 @@
     if (!ref)
     {
         EnterCriticalSection(&sample->parent->cs);
-
-        if (sample->pending)
-            remove_queued_update(sample);
-
-        while (sample->media_sample_refcount)
-            SleepConditionVariableCS(&sample->parent->allocator_cv, &sample->parent->cs, INFINITE);
-
         --sample->parent->sample_refs;
         LeaveCriticalSection(&sample->parent->cs);
 
@@ -1967,7 +1633,7 @@
         LeaveCriticalSection(&sample->parent->cs);
         return MS_S_ENDOFSTREAM;
     }
-    if (sample->pending || sample->media_sample_refcount)
+    if (sample->busy)
     {
         LeaveCriticalSection(&sample->parent->cs);
         return MS_E_BUSY;
@@ -1976,11 +1642,10 @@
     sample->continuous_update = (flags & SSUPDATE_ASYNC) && (flags & SSUPDATE_CONTINUOUS);
 
     sample->update_hr = MS_S_NOUPDATE;
-    sample->pending = true;
+    sample->busy = TRUE;
     sample->external_event = event;
     list_add_tail(&sample->parent->update_queue, &sample->entry);
     WakeConditionVariable(&sample->parent->update_queued_cv);
-    WakeConditionVariable(&sample->parent->allocator_cv);
 
     if ((flags & SSUPDATE_ASYNC) || event)
     {
@@ -1988,7 +1653,7 @@
         return MS_S_PENDING;
     }
 
-    while (sample->pending || sample->media_sample_refcount)
+    while (sample->busy)
         SleepConditionVariableCS(&sample->update_cv, &sample->parent->cs, INFINITE);
 
     LeaveCriticalSection(&sample->parent->cs);
@@ -2005,19 +1670,18 @@
 
     EnterCriticalSection(&sample->parent->cs);
 
-    if (sample->pending || sample->media_sample_refcount)
+    if (sample->busy)
     {
         if (flags & (COMPSTAT_NOUPDATEOK | COMPSTAT_ABORT))
         {
-            if (!sample->media_sample_refcount)
-                remove_queued_update(sample);
+            remove_queued_update(sample);
         }
         else if (flags & COMPSTAT_WAIT)
         {
             DWORD start_time = GetTickCount();
             DWORD elapsed = 0;
             sample->continuous_update = FALSE;
-            while ((sample->pending || sample->media_sample_refcount) && elapsed < milliseconds)
+            while (sample->busy && elapsed < milliseconds)
             {
                 DWORD sleep_time = milliseconds - elapsed;
                 if (!SleepConditionVariableCS(&sample->update_cv, &sample->parent->cs, sleep_time))
@@ -2027,10 +1691,7 @@
         }
     }
 
-    if (sample->pending || sample->media_sample_refcount)
-        hr = MS_S_PENDING;
-    else
-        hr = sample->update_hr;
+    hr = sample->busy ? MS_S_PENDING : sample->update_hr;
 
     LeaveCriticalSection(&sample->parent->cs);
 
@@ -2082,306 +1743,6 @@
     ddraw_sample_SetRect
 };
 
-static struct ddraw_sample *impl_from_IMediaSample(IMediaSample *iface)
-{
-    return CONTAINING_RECORD(iface, struct ddraw_sample, IMediaSample_iface);
-}
-
-static HRESULT WINAPI media_sample_QueryInterface(IMediaSample *iface, REFIID iid, void **out)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, iid %s, out %p.\n", sample, debugstr_guid(iid), out);
-
-    if (IsEqualIID(iid, &IID_IUnknown) || IsEqualIID(iid, &IID_IMediaSample))
-    {
-        IMediaSample_AddRef(iface);
-        *out = iface;
-        return S_OK;
-    }
-
-    WARN("%s not implemented, returning E_NOINTERFACE.\n", debugstr_guid(iid));
-    *out = NULL;
-    return E_NOINTERFACE;
-}
-
-static ULONG WINAPI media_sample_AddRef(IMediaSample *iface)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-    ULONG refcount;
-
-    EnterCriticalSection(&sample->parent->cs);
-    refcount = ++sample->media_sample_refcount;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    TRACE("%p increasing refcount to %lu.\n", sample, refcount);
-
-    return refcount;
-}
-
-static ULONG WINAPI media_sample_Release(IMediaSample *iface)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-    ULONG refcount;
-
-    EnterCriticalSection(&sample->parent->cs);
-
-    /* This is not InterlockedDecrement(), because it's used in GetBuffer()
-     * among other functions and therefore needs to be protected by the CS
-     * anyway. */
-    refcount = --sample->media_sample_refcount;
-
-    TRACE("%p decreasing refcount to %lu.\n", sample, refcount);
-
-    if (!refcount)
-    {
-        IDirectDrawSurface_Unlock(sample->surface, NULL);
-
-        WakeConditionVariable(&sample->update_cv);
-
-        /* This sample is not released back to the pool if it's no longer
-         * pending an update.
-         *
-         * Use WakeAll, even though we're only releasing one sample, because we
-         * also potentially need to wake up a thread stuck in
-         * IDirectDrawStreamSample::Release().
-         * This is arguably wasteful, but in practice we're unlikely to have
-         * more than one thread in GetBuffer() at a time anyway. */
-        if (sample->pending)
-            WakeAllConditionVariable(&sample->parent->allocator_cv);
-    }
-
-    LeaveCriticalSection(&sample->parent->cs);
-    return refcount;
-}
-
-static unsigned int get_sample_size(struct ddraw_sample *sample)
-{
-    return sample->surface_desc.lPitch * sample->surface_desc.dwHeight;
-}
-
-static HRESULT WINAPI media_sample_GetPointer(IMediaSample *iface, BYTE **data)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, data %p.\n", sample, data);
-
-    *data = sample->surface_desc.lpSurface;
-    return S_OK;
-}
-
-static LONG WINAPI media_sample_GetSize(IMediaSample *iface)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p.\n", sample);
-
-    return get_sample_size(sample);
-}
-
-static HRESULT WINAPI media_sample_GetTime(IMediaSample *iface, REFERENCE_TIME *start, REFERENCE_TIME *end)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, start %p, end %p.\n", sample, start, end);
-
-    EnterCriticalSection(&sample->parent->cs);
-    *start = sample->start_time;
-    *end = sample->end_time;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    return S_OK;
-}
-
-static HRESULT WINAPI media_sample_SetTime(IMediaSample *iface, REFERENCE_TIME *start, REFERENCE_TIME *end)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, start %p, end %p.\n", sample, start, end);
-
-    EnterCriticalSection(&sample->parent->cs);
-    if (start)
-        sample->start_time = *start;
-    if (end)
-        sample->end_time = *end;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    return S_OK;
-}
-
-static HRESULT WINAPI media_sample_IsSyncPoint(IMediaSample *iface)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-    HRESULT hr;
-
-    TRACE("sample %p.\n", sample);
-
-    EnterCriticalSection(&sample->parent->cs);
-    hr = sample->sync_point ? S_OK : S_FALSE;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    return hr;
-}
-
-static HRESULT WINAPI media_sample_SetSyncPoint(IMediaSample *iface, BOOL sync_point)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, sync_point %d.\n", sample, sync_point);
-
-    EnterCriticalSection(&sample->parent->cs);
-    sample->sync_point = sync_point;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    return S_OK;
-}
-
-static HRESULT WINAPI media_sample_IsPreroll(IMediaSample *iface)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-    HRESULT hr;
-
-    TRACE("sample %p.\n", sample);
-
-    EnterCriticalSection(&sample->parent->cs);
-    hr = sample->preroll ? S_OK : S_FALSE;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    return hr;
-}
-
-static HRESULT WINAPI media_sample_SetPreroll(IMediaSample *iface, BOOL preroll)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, preroll %d.\n", sample, preroll);
-
-    EnterCriticalSection(&sample->parent->cs);
-    sample->preroll = preroll;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    return S_OK;
-}
-
-static LONG WINAPI media_sample_GetActualDataLength(IMediaSample *iface)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p.\n", sample);
-
-    return get_sample_size(sample);
-}
-
-static HRESULT WINAPI media_sample_SetActualDataLength(IMediaSample *iface, LONG length)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, length %ld.\n", sample, length);
-
-    return (length == get_sample_size(sample) ? S_OK : E_FAIL);
-}
-
-static HRESULT WINAPI media_sample_GetMediaType(IMediaSample *iface, AM_MEDIA_TYPE **ret_mt)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-    VIDEOINFOHEADER *video_info;
-
-    TRACE("sample %p, ret_mt %p.\n", sample, ret_mt);
-
-    /* Note that this usually matches the media type we pass to QueryAccept(),
-     * but not if there's a sub-rect.
-     * That's amstream just breaking the DirectShow rules.
-     * The type we pass to QueryAccept() just uses the size of the sub-rect for
-     * everything. The type we return from GetMediaType() uses the size of the
-     * surface for everything except rcSource/rcTarget. */
-    if (!(*ret_mt = CoTaskMemAlloc(sizeof(AM_MEDIA_TYPE))))
-        return E_OUTOFMEMORY;
-    set_mt_from_desc(*ret_mt, &sample->surface_desc, sample->surface_desc.lPitch);
-    video_info = (VIDEOINFOHEADER *)(*ret_mt)->pbFormat;
-    SetRect(&video_info->rcSource, 0, 0, sample->rect.right - sample->rect.left,
-            sample->rect.bottom - sample->rect.top);
-    video_info->rcTarget = sample->rect;
-    return S_OK;
-}
-
-static HRESULT WINAPI media_sample_SetMediaType(IMediaSample *iface, AM_MEDIA_TYPE *mt)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    FIXME("sample %p, mt %p, stub!\n", sample, mt);
-    strmbase_dump_media_type(mt);
-
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI media_sample_IsDiscontinuity(IMediaSample *iface)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-    HRESULT hr;
-
-    TRACE("sample %p.\n", sample);
-
-    EnterCriticalSection(&sample->parent->cs);
-    hr = sample->discontinuity ? S_OK : S_FALSE;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    return hr;
-}
-
-static HRESULT WINAPI media_sample_SetDiscontinuity(IMediaSample *iface, BOOL discontinuity)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, discontinuity %d.\n", sample, discontinuity);
-
-    EnterCriticalSection(&sample->parent->cs);
-    sample->discontinuity = discontinuity;
-    LeaveCriticalSection(&sample->parent->cs);
-
-    return S_OK;
-}
-
-static HRESULT WINAPI media_sample_GetMediaTime(IMediaSample *iface, LONGLONG *start, LONGLONG *end)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, start %p, end %p, not implemented.\n", sample, start, end);
-
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI media_sample_SetMediaTime(IMediaSample *iface, LONGLONG *start, LONGLONG *end)
-{
-    struct ddraw_sample *sample = impl_from_IMediaSample(iface);
-
-    TRACE("sample %p, start %p, end %p, not implemented.\n", sample, start, end);
-
-    return E_NOTIMPL;
-}
-
-static const struct IMediaSampleVtbl media_sample_vtbl =
-{
-    media_sample_QueryInterface,
-    media_sample_AddRef,
-    media_sample_Release,
-    media_sample_GetPointer,
-    media_sample_GetSize,
-    media_sample_GetTime,
-    media_sample_SetTime,
-    media_sample_IsSyncPoint,
-    media_sample_SetSyncPoint,
-    media_sample_IsPreroll,
-    media_sample_SetPreroll,
-    media_sample_GetActualDataLength,
-    media_sample_SetActualDataLength,
-    media_sample_GetMediaType,
-    media_sample_SetMediaType,
-    media_sample_IsDiscontinuity,
-    media_sample_SetDiscontinuity,
-    media_sample_GetMediaTime,
-    media_sample_SetMediaTime,
-};
-
 static HRESULT ddrawstreamsample_create(struct ddraw_stream *parent, IDirectDrawSurface *surface,
     const RECT *rect, IDirectDrawStreamSample **ddraw_stream_sample)
 {
@@ -2391,27 +1752,10 @@
 
     TRACE("(%p)\n", ddraw_stream_sample);
 
-    if (surface)
-    {
-        desc.dwSize = sizeof(desc);
-        if (FAILED(hr = IDirectDrawSurface_GetSurfaceDesc(surface, &desc)))
-            return hr;
-
-        if (rect)
-        {
-            desc.dwWidth = rect->right - rect->left;
-            desc.dwHeight = rect->bottom - rect->top;
-        }
-
-        if (FAILED(hr = IDirectDrawMediaStream_SetFormat(&parent->IDirectDrawMediaStream_iface, &desc, NULL)))
-            return hr;
-    }
-
     if (!(object = calloc(1, sizeof(*object))))
         return E_OUTOFMEMORY;
 
     object->IDirectDrawStreamSample_iface.lpVtbl = &DirectDrawStreamSample_Vtbl;
-    object->IMediaSample_iface.lpVtbl = &media_sample_vtbl;
     object->ref = 1;
     object->parent = parent;
     object->mmstream = parent->parent;
@@ -2425,11 +1769,6 @@
     {
         object->surface = surface;
         IDirectDrawSurface_AddRef(surface);
-
-        if (rect)
-            object->rect = *rect;
-        else
-            SetRect(&object->rect, 0, 0, desc.dwWidth, desc.dwHeight);
     }
     else
     {
@@ -2471,10 +1810,34 @@
             IDirectDrawStreamSample_Release(&object->IDirectDrawStreamSample_iface);
             return hr;
         }
-
+    }
+
+    desc.dwSize = sizeof(desc);
+    hr = IDirectDrawSurface_GetSurfaceDesc(object->surface, &desc);
+    if (FAILED(hr))
+    {
+        IDirectDrawStreamSample_Release(&object->IDirectDrawStreamSample_iface);
+        return hr;
+    }
+
+    if (rect)
+    {
+        object->rect = *rect;
+        desc.dwWidth = rect->right - rect->left;
+        desc.dwHeight = rect->bottom - rect->top;
+    }
+    else
+    {
         SetRect(&object->rect, 0, 0, desc.dwWidth, desc.dwHeight);
     }
 
+    hr = IDirectDrawMediaStream_SetFormat(&parent->IDirectDrawMediaStream_iface, &desc, NULL);
+    if (FAILED(hr))
+    {
+        IDirectDrawStreamSample_Release(&object->IDirectDrawStreamSample_iface);
+        return hr;
+    }
+
     *ddraw_stream_sample = &object->IDirectDrawStreamSample_iface;
 
     return S_OK;
