diff --git a/dlls/d2d1/geometry.c b/dlls/d2d1/geometry.c
--- a/dlls/d2d1/geometry.c
+++ b/dlls/d2d1/geometry.c
@@ -276,30 +276,50 @@
 static void d2d_fp_fast_expansion_sum_zeroelim(float *out, size_t *out_len,
         const float *a, size_t a_len, const float *b, size_t b_len)
 {
-    float sum[2], q;
+    float sum[2], q, a_curr, b_curr;
     size_t a_idx, b_idx, out_idx;
 
+    a_curr = a[0];
+    b_curr = b[0];
     a_idx = b_idx = 0;
-    if ((b[b_idx] > a[a_idx]) == (b[b_idx] > -a[a_idx]))
-        q = a[a_idx++];
+    if ((b_curr > a_curr) == (b_curr > -a_curr))
+    {
+        q = a_curr;
+        a_curr = a[++a_idx];
+    }
     else
-        q = b[b_idx++];
+    {
+        q = b_curr;
+        b_curr = b[++b_idx];
+    }
     out_idx = 0;
     if (a_idx < a_len && b_idx < b_len)
     {
-        if ((b[b_idx] > a[a_idx]) == (b[b_idx] > -a[a_idx]))
-            d2d_fp_fast_two_sum(sum, a[a_idx++], q);
+        if ((b_curr > a_curr) == (b_curr > -a_curr))
+        {
+            d2d_fp_fast_two_sum(sum, a_curr, q);
+            a_curr = a[++a_idx];
+        }
         else
-            d2d_fp_fast_two_sum(sum, b[b_idx++], q);
+        {
+            d2d_fp_fast_two_sum(sum, b_curr, q);
+            b_curr = b[++b_idx];
+        }
         if (sum[0] != 0.0f)
             out[out_idx++] = sum[0];
         q = sum[1];
         while (a_idx < a_len && b_idx < b_len)
         {
-            if ((b[b_idx] > a[a_idx]) == (b[b_idx] > -a[a_idx]))
-                d2d_fp_two_sum(sum, q, a[a_idx++]);
+            if ((b_curr > a_curr) == (b_curr > -a_curr))
+            {
+                d2d_fp_two_sum(sum, q, a_curr);
+                a_curr = a[++a_idx];
+            }
             else
-                d2d_fp_two_sum(sum, q, b[b_idx++]);
+            {
+                d2d_fp_two_sum(sum, q, b_curr);
+                b_curr = b[++b_idx];
+            }
             if (sum[0] != 0.0f)
                 out[out_idx++] = sum[0];
             q = sum[1];
@@ -307,14 +327,16 @@
     }
     while (a_idx < a_len)
     {
-        d2d_fp_two_sum(sum, q, a[a_idx++]);
+        d2d_fp_two_sum(sum, q, a_curr);
+        a_curr = a[++a_idx];
         if (sum[0] != 0.0f)
             out[out_idx++] = sum[0];
         q = sum[1];
     }
     while (b_idx < b_len)
     {
-        d2d_fp_two_sum(sum, q, b[b_idx++]);
+        d2d_fp_two_sum(sum, q, b_curr);
+        b_curr = b[++b_idx];
         if (sum[0] != 0.0f)
             out[out_idx++] = sum[0];
         q = sum[1];
@@ -822,33 +844,33 @@
     return TRUE;
 }
 
-static bool d2d_figure_add_vertex(struct d2d_figure *figure, D2D1_POINT_2F vertex)
+static BOOL d2d_figure_add_vertex(struct d2d_figure *figure, D2D1_POINT_2F vertex)
 {
     size_t last = figure->vertex_count - 1;
 
     if (figure->vertex_count && figure->vertex_types[last] == D2D_VERTEX_TYPE_LINE
             && !memcmp(&figure->vertices[last], &vertex, sizeof(vertex)))
-        return true;
+        return TRUE;
 
     if (!d2d_array_reserve((void **)&figure->vertices, &figure->vertices_size,
             figure->vertex_count + 1, sizeof(*figure->vertices)))
     {
         ERR("Failed to grow vertices array.\n");
-        return false;
+        return FALSE;
     }
 
     if (!d2d_array_reserve((void **)&figure->vertex_types, &figure->vertex_types_size,
             figure->vertex_count + 1, sizeof(*figure->vertex_types)))
     {
         ERR("Failed to grow vertex types array.\n");
-        return false;
+        return FALSE;
     }
 
     figure->vertices[figure->vertex_count] = vertex;
     figure->vertex_types[figure->vertex_count] = D2D_VERTEX_TYPE_NONE;
     d2d_rect_expand(&figure->bounds, &vertex);
     ++figure->vertex_count;
-    return true;
+    return TRUE;
 }
 
 static BOOL d2d_figure_insert_bezier_controls(struct d2d_figure *figure,
@@ -898,86 +920,6 @@
     figure->original_bezier_control_count += count;
 
     return TRUE;
-}
-
-static bool d2d_figure_begin(struct d2d_figure *figure, D2D1_POINT_2F start_point,
-        D2D1_FIGURE_BEGIN figure_begin)
-{
-    if (figure_begin == D2D1_FIGURE_BEGIN_HOLLOW)
-        figure->flags |= D2D_FIGURE_FLAG_HOLLOW;
-
-    return d2d_figure_add_vertex(figure, start_point);
-}
-
-static void d2d_figure_end(struct d2d_figure *figure, D2D1_FIGURE_END figure_end)
-{
-    if (memcmp(&figure->vertices[0], &figure->vertices[figure->vertex_count - 1], sizeof(*figure->vertices)))
-        figure->vertex_types[figure->vertex_count - 1] = D2D_VERTEX_TYPE_LINE;
-    else
-        figure->vertex_types[figure->vertex_count - 1] = D2D_VERTEX_TYPE_END;
-    if (figure_end == D2D1_FIGURE_END_CLOSED)
-        figure->flags |= D2D_FIGURE_FLAG_CLOSED;
-}
-
-static bool d2d_figure_add_beziers(struct d2d_figure *figure, const D2D1_BEZIER_SEGMENT *beziers,
-        UINT32 count)
-{
-    D2D1_POINT_2F p;
-    unsigned int i;
-
-    for (i = 0; i < count; ++i)
-    {
-        D2D1_RECT_F bezier_bounds;
-
-        if (!d2d_figure_add_original_bezier_controls(figure, 1, &beziers[i].point1)
-                || !d2d_figure_add_original_bezier_controls(figure, 1, &beziers[i].point2))
-        {
-            return false;
-        }
-
-        /* FIXME: This tries to approximate a cubic Bézier with a quadratic one. */
-        p.x = (beziers[i].point1.x + beziers[i].point2.x) * 0.75f;
-        p.y = (beziers[i].point1.y + beziers[i].point2.y) * 0.75f;
-        p.x -= (figure->vertices[figure->vertex_count - 1].x + beziers[i].point3.x) * 0.25f;
-        p.y -= (figure->vertices[figure->vertex_count - 1].y + beziers[i].point3.y) * 0.25f;
-        figure->vertex_types[figure->vertex_count - 1] = D2D_VERTEX_TYPE_BEZIER;
-
-        d2d_rect_get_bezier_bounds(&bezier_bounds, &figure->vertices[figure->vertex_count - 1],
-                &p, &beziers[i].point3);
-
-        if (!d2d_figure_add_bezier_controls(figure, 1, &p))
-            return false;
-
-        if (!d2d_figure_add_vertex(figure, beziers[i].point3))
-            return false;
-
-        d2d_rect_union(&figure->bounds, &bezier_bounds);
-    }
-
-    return true;
-}
-
-static bool d2d_figure_add_lines(struct d2d_figure *figure, const D2D1_POINT_2F *points,
-        UINT32 count)
-{
-    unsigned int i;
-
-    for (i = 0; i < count; ++i)
-    {
-        figure->vertex_types[figure->vertex_count - 1] = D2D_VERTEX_TYPE_LINE;
-        if (!d2d_figure_add_vertex(figure, points[i]))
-            return false;
-    }
-
-    return true;
-}
-
-static void d2d_figure_cleanup(struct d2d_figure *figure)
-{
-    free(figure->original_bezier_controls);
-    free(figure->bezier_controls);
-    free(figure->vertices);
-    memset(figure, 0, sizeof(*figure));
 }
 
 static void d2d_cdt_edge_rot(struct d2d_cdt_edge_ref *dst, const struct d2d_cdt_edge_ref *src)
@@ -2851,26 +2793,9 @@
     geometry->u.path.fill_mode = mode;
 }
 
-static void d2d_geometry_set_error(struct d2d_geometry *geometry, HRESULT code)
-{
-    if (geometry->u.path.state == D2D_GEOMETRY_STATE_ERROR)
-        return;
-
-    geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
-    geometry->u.path.code = code;
-}
-
 static void STDMETHODCALLTYPE d2d_geometry_sink_SetSegmentFlags(ID2D1GeometrySink *iface, D2D1_PATH_SEGMENT flags)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1GeometrySink(iface);
-
     TRACE("iface %p, flags %#x.\n", iface, flags);
-
-    if (flags & ~(D2D1_PATH_SEGMENT_FORCE_UNSTROKED | D2D1_PATH_SEGMENT_FORCE_ROUND_LINE_JOIN))
-    {
-        d2d_geometry_set_error(geometry, E_INVALIDARG);
-        return;
-    }
 
     if (flags != D2D1_PATH_SEGMENT_NONE)
         FIXME("Ignoring flags %#x.\n", flags);
@@ -2887,21 +2812,25 @@
 
     if (geometry->u.path.state != D2D_GEOMETRY_STATE_OPEN)
     {
-        d2d_geometry_set_error(geometry, D2DERR_WRONG_STATE);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
     if (!d2d_path_geometry_add_figure(geometry))
     {
         ERR("Failed to add figure.\n");
-        d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
     figure = &geometry->u.path.figures[geometry->u.path.figure_count - 1];
-    if (!d2d_figure_begin(figure, start_point, figure_begin))
-    {
-        d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
+    if (figure_begin == D2D1_FIGURE_BEGIN_HOLLOW)
+        figure->flags |= D2D_FIGURE_FLAG_HOLLOW;
+
+    if (!d2d_figure_add_vertex(figure, start_point))
+    {
+        ERR("Failed to add vertex.\n");
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
@@ -2913,20 +2842,24 @@
 {
     struct d2d_geometry *geometry = impl_from_ID2D1GeometrySink(iface);
     struct d2d_figure *figure = &geometry->u.path.figures[geometry->u.path.figure_count - 1];
+    unsigned int i;
 
     TRACE("iface %p, points %p, count %u.\n", iface, points, count);
 
     if (geometry->u.path.state != D2D_GEOMETRY_STATE_FIGURE)
     {
-        d2d_geometry_set_error(geometry, D2DERR_WRONG_STATE);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
-    if (!d2d_figure_add_lines(figure, points, count))
-    {
-        ERR("Failed to add vertex.\n");
-        d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
-        return;
+    for (i = 0; i < count; ++i)
+    {
+        figure->vertex_types[figure->vertex_count - 1] = D2D_VERTEX_TYPE_LINE;
+        if (!d2d_figure_add_vertex(figure, points[i]))
+        {
+            ERR("Failed to add vertex.\n");
+            return;
+        }
     }
 
     geometry->u.path.segment_count += count;
@@ -2937,20 +2870,54 @@
 {
     struct d2d_geometry *geometry = impl_from_ID2D1GeometrySink(iface);
     struct d2d_figure *figure = &geometry->u.path.figures[geometry->u.path.figure_count - 1];
+    D2D1_POINT_2F p;
+    unsigned int i;
 
     TRACE("iface %p, beziers %p, count %u.\n", iface, beziers, count);
 
     if (geometry->u.path.state != D2D_GEOMETRY_STATE_FIGURE)
     {
-        d2d_geometry_set_error(geometry, D2DERR_WRONG_STATE);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
-    if (!d2d_figure_add_beziers(figure, beziers, count))
-    {
-        ERR("Failed to add Bézier curves.\n");
-        d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
-        return;
+    for (i = 0; i < count; ++i)
+    {
+        D2D1_RECT_F bezier_bounds;
+
+        if (!d2d_figure_add_original_bezier_controls(figure, 1, &beziers[i].point1)
+                || !d2d_figure_add_original_bezier_controls(figure, 1, &beziers[i].point2))
+        {
+            ERR("Failed to add cubic Bézier controls.\n");
+            geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
+            return;
+        }
+
+        /* FIXME: This tries to approximate a cubic Bézier with a quadratic one. */
+        p.x = (beziers[i].point1.x + beziers[i].point2.x) * 0.75f;
+        p.y = (beziers[i].point1.y + beziers[i].point2.y) * 0.75f;
+        p.x -= (figure->vertices[figure->vertex_count - 1].x + beziers[i].point3.x) * 0.25f;
+        p.y -= (figure->vertices[figure->vertex_count - 1].y + beziers[i].point3.y) * 0.25f;
+        figure->vertex_types[figure->vertex_count - 1] = D2D_VERTEX_TYPE_BEZIER;
+
+        d2d_rect_get_bezier_bounds(&bezier_bounds, &figure->vertices[figure->vertex_count - 1],
+                &p, &beziers[i].point3);
+
+        if (!d2d_figure_add_bezier_controls(figure, 1, &p))
+        {
+            ERR("Failed to add bezier control.\n");
+            geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
+            return;
+        }
+
+        if (!d2d_figure_add_vertex(figure, beziers[i].point3))
+        {
+            ERR("Failed to add bezier vertex.\n");
+            geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
+            return;
+        }
+
+        d2d_rect_union(&figure->bounds, &bezier_bounds);
     }
 
     geometry->u.path.segment_count += count;
@@ -2965,20 +2932,25 @@
 
     if (geometry->u.path.state != D2D_GEOMETRY_STATE_FIGURE)
     {
-        d2d_geometry_set_error(geometry, D2DERR_WRONG_STATE);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
     figure = &geometry->u.path.figures[geometry->u.path.figure_count - 1];
-    d2d_figure_end(figure, figure_end);
-
+    if (memcmp(&figure->vertices[0], &figure->vertices[figure->vertex_count - 1], sizeof(*figure->vertices)))
+        figure->vertex_types[figure->vertex_count - 1] = D2D_VERTEX_TYPE_LINE;
+    else
+        figure->vertex_types[figure->vertex_count - 1] = D2D_VERTEX_TYPE_END;
     if (figure_end == D2D1_FIGURE_END_CLOSED)
+    {
         ++geometry->u.path.segment_count;
+        figure->flags |= D2D_FIGURE_FLAG_CLOSED;
+    }
 
     if (!d2d_geometry_add_figure_outline(geometry, figure, figure_end))
     {
         ERR("Failed to add figure outline.\n");
-        d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
@@ -2993,8 +2965,11 @@
         return;
 
     for (i = 0; i < geometry->u.path.figure_count; ++i)
-        d2d_figure_cleanup(&geometry->u.path.figures[i]);
-
+    {
+        free(geometry->u.path.figures[i].original_bezier_controls);
+        free(geometry->u.path.figures[i].bezier_controls);
+        free(geometry->u.path.figures[i].vertices);
+    }
     free(geometry->u.path.figures);
     geometry->u.path.figures = NULL;
     geometry->u.path.figures_size = 0;
@@ -3253,15 +3228,12 @@
 
     TRACE("iface %p.\n", iface);
 
-    if (geometry->u.path.state == D2D_GEOMETRY_STATE_CLOSED)
+    if (geometry->u.path.state != D2D_GEOMETRY_STATE_OPEN)
+    {
+        if (geometry->u.path.state != D2D_GEOMETRY_STATE_CLOSED)
+            geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return D2DERR_WRONG_STATE;
-
-    if (geometry->u.path.state != D2D_GEOMETRY_STATE_OPEN)
-        d2d_geometry_set_error(geometry, D2DERR_WRONG_STATE);
-
-    if (geometry->u.path.state == D2D_GEOMETRY_STATE_ERROR)
-        return geometry->u.path.code;
-
+    }
     geometry->u.path.state = D2D_GEOMETRY_STATE_CLOSED;
 
     if (!d2d_geometry_intersect_self(geometry))
@@ -3278,7 +3250,7 @@
         geometry->fill.bezier_vertices = NULL;
         geometry->fill.bezier_vertex_count = 0;
         d2d_path_geometry_free_figures(geometry);
-        d2d_geometry_set_error(geometry, hr);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
     }
     return hr;
 }
@@ -3316,7 +3288,7 @@
 
     if (geometry->u.path.state != D2D_GEOMETRY_STATE_FIGURE)
     {
-        d2d_geometry_set_error(geometry, D2DERR_WRONG_STATE);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
@@ -3331,7 +3303,7 @@
         if (!d2d_figure_add_original_bezier_controls(figure, 2, p))
         {
             ERR("Failed to add cubic Bézier controls.\n");
-            d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
+            geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
             return;
         }
 
@@ -3342,14 +3314,14 @@
         if (!d2d_figure_add_bezier_controls(figure, 1, &beziers[i].point1))
         {
             ERR("Failed to add bezier.\n");
-            d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
+            geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
             return;
         }
 
         if (!d2d_figure_add_vertex(figure, beziers[i].point2))
         {
             ERR("Failed to add bezier vertex.\n");
-            d2d_geometry_set_error(geometry, E_OUTOFMEMORY);
+            geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
             return;
         }
 
@@ -3367,7 +3339,7 @@
 
     if (geometry->u.path.state != D2D_GEOMETRY_STATE_FIGURE)
     {
-        d2d_geometry_set_error(geometry, D2DERR_WRONG_STATE);
+        geometry->u.path.state = D2D_GEOMETRY_STATE_ERROR;
         return;
     }
 
@@ -3753,88 +3725,17 @@
     ID2D1SimplifiedGeometrySink_SetSegmentFlags(sink, D2D1_PATH_SEGMENT_NONE);
 }
 
-static void d2d_figure_simplify(const struct d2d_figure *figure,
-        D2D1_GEOMETRY_SIMPLIFICATION_OPTION option, const D2D1_MATRIX_3X2_F *transform,
-        float tolerance, ID2D1SimplifiedGeometrySink *sink)
-{
-    enum d2d_vertex_type type = D2D_VERTEX_TYPE_NONE;
-    D2D1_FIGURE_BEGIN begin;
-    D2D1_BEZIER_SEGMENT b;
-    size_t i, bezier_idx;
-    D2D1_FIGURE_END end;
-    D2D1_POINT_2F p;
-
-    for (i = 0; i < figure->vertex_count; ++i)
-    {
-        if (figure->vertex_types[i] == D2D_VERTEX_TYPE_NONE)
-            continue;
-
-        p = figure->vertices[i];
-        if (transform)
-            d2d_point_transform(&p, transform, p.x, p.y);
-        begin = figure->flags & D2D_FIGURE_FLAG_HOLLOW ? D2D1_FIGURE_BEGIN_HOLLOW : D2D1_FIGURE_BEGIN_FILLED;
-        ID2D1SimplifiedGeometrySink_BeginFigure(sink, p, begin);
-        type = figure->vertex_types[i];
-        break;
-    }
-
-    for (bezier_idx = 0, ++i; i < figure->vertex_count; ++i)
-    {
-        enum d2d_vertex_type next_type;
-
-        if ((next_type = figure->vertex_types[i]) == D2D_VERTEX_TYPE_NONE
-                || d2d_vertex_type_is_split_bezier(next_type))
-            continue;
-
-        switch (type)
-        {
-            case D2D_VERTEX_TYPE_LINE:
-                p = figure->vertices[i];
-                if (transform)
-                    d2d_point_transform(&p, transform, p.x, p.y);
-                ID2D1SimplifiedGeometrySink_AddLines(sink, &p, 1);
-                break;
-
-            case D2D_VERTEX_TYPE_BEZIER:
-                b.point1 = figure->original_bezier_controls[bezier_idx++];
-                b.point2 = figure->original_bezier_controls[bezier_idx++];
-                b.point3 = figure->vertices[i];
-                if (transform)
-                {
-                    d2d_point_transform(&b.point1, transform, b.point1.x, b.point1.y);
-                    d2d_point_transform(&b.point2, transform, b.point2.x, b.point2.y);
-                    d2d_point_transform(&b.point3, transform, b.point3.x, b.point3.y);
-                }
-
-                if (option == D2D1_GEOMETRY_SIMPLIFICATION_OPTION_LINES)
-                    d2d_geometry_flatten_cubic(sink, &p, &b, tolerance);
-                else
-                    ID2D1SimplifiedGeometrySink_AddBeziers(sink, &b, 1);
-                p = b.point3;
-                break;
-
-            default:
-                FIXME("Unhandled vertex type %#x.\n", type);
-                p = figure->vertices[i];
-                if (transform)
-                    d2d_point_transform(&p, transform, p.x, p.y);
-                ID2D1SimplifiedGeometrySink_AddLines(sink, &p, 1);
-                break;
-        }
-
-        type = next_type;
-    }
-
-    end = figure->flags & D2D_FIGURE_FLAG_CLOSED ? D2D1_FIGURE_END_CLOSED : D2D1_FIGURE_END_OPEN;
-    ID2D1SimplifiedGeometrySink_EndFigure(sink, end);
-}
-
 static HRESULT STDMETHODCALLTYPE d2d_path_geometry_Simplify(ID2D1PathGeometry1 *iface,
         D2D1_GEOMETRY_SIMPLIFICATION_OPTION option, const D2D1_MATRIX_3X2_F *transform, float tolerance,
         ID2D1SimplifiedGeometrySink *sink)
 {
     struct d2d_geometry *geometry = impl_from_ID2D1PathGeometry1(iface);
-    size_t i;
+    enum d2d_vertex_type type = D2D_VERTEX_TYPE_NONE;
+    unsigned int i, j, bezier_idx;
+    D2D1_FIGURE_BEGIN begin;
+    D2D1_BEZIER_SEGMENT b;
+    D2D1_FIGURE_END end;
+    D2D1_POINT_2F p;
 
     TRACE("iface %p, option %#x, transform %p, tolerance %.8e, sink %p.\n",
             iface, option, transform, tolerance, sink);
@@ -3844,128 +3745,80 @@
     {
         const struct d2d_figure *figure = &geometry->u.path.figures[i];
 
-        d2d_figure_simplify(figure, option, transform, tolerance, sink);
+        for (j = 0; j < figure->vertex_count; ++j)
+        {
+            if (figure->vertex_types[j] == D2D_VERTEX_TYPE_NONE)
+                continue;
+
+            p = figure->vertices[j];
+            if (transform)
+                d2d_point_transform(&p, transform, p.x, p.y);
+            begin = figure->flags & D2D_FIGURE_FLAG_HOLLOW ? D2D1_FIGURE_BEGIN_HOLLOW : D2D1_FIGURE_BEGIN_FILLED;
+            ID2D1SimplifiedGeometrySink_BeginFigure(sink, p, begin);
+            type = figure->vertex_types[j];
+            break;
+        }
+
+        for (bezier_idx = 0, ++j; j < figure->vertex_count; ++j)
+        {
+            enum d2d_vertex_type next_type;
+
+            if ((next_type = figure->vertex_types[j]) == D2D_VERTEX_TYPE_NONE
+                    || d2d_vertex_type_is_split_bezier(next_type))
+                continue;
+
+            switch (type)
+            {
+                case D2D_VERTEX_TYPE_LINE:
+                    p = figure->vertices[j];
+                    if (transform)
+                        d2d_point_transform(&p, transform, p.x, p.y);
+                    ID2D1SimplifiedGeometrySink_AddLines(sink, &p, 1);
+                    break;
+
+                case D2D_VERTEX_TYPE_BEZIER:
+                    b.point1 = figure->original_bezier_controls[bezier_idx++];
+                    b.point2 = figure->original_bezier_controls[bezier_idx++];
+                    b.point3 = figure->vertices[j];
+                    if (transform)
+                    {
+                        d2d_point_transform(&b.point1, transform, b.point1.x, b.point1.y);
+                        d2d_point_transform(&b.point2, transform, b.point2.x, b.point2.y);
+                        d2d_point_transform(&b.point3, transform, b.point3.x, b.point3.y);
+                    }
+
+                    if (option == D2D1_GEOMETRY_SIMPLIFICATION_OPTION_LINES)
+                        d2d_geometry_flatten_cubic(sink, &p, &b, tolerance);
+                    else
+                        ID2D1SimplifiedGeometrySink_AddBeziers(sink, &b, 1);
+                    p = b.point3;
+                    break;
+
+                default:
+                    FIXME("Unhandled vertex type %#x.\n", type);
+                    p = figure->vertices[j];
+                    if (transform)
+                        d2d_point_transform(&p, transform, p.x, p.y);
+                    ID2D1SimplifiedGeometrySink_AddLines(sink, &p, 1);
+                    break;
+            }
+
+            type = next_type;
+        }
+
+        end = figure->flags & D2D_FIGURE_FLAG_CLOSED ? D2D1_FIGURE_END_CLOSED : D2D1_FIGURE_END_OPEN;
+        ID2D1SimplifiedGeometrySink_EndFigure(sink, end);
     }
 
     return S_OK;
-}
-
-static HRESULT d2d_geometry_get_simplified(ID2D1Geometry *geometry, const D2D1_MATRIX_3X2_F *transform,
-        float tolerance, ID2D1PathGeometry **ret)
-{
-    ID2D1PathGeometry *path_geometry = NULL;
-    ID2D1GeometrySink *geometry_sink = NULL;
-    ID2D1Factory *factory;
-    HRESULT hr;
-
-    *ret = NULL;
-
-    ID2D1Geometry_GetFactory(geometry, &factory);
-
-    hr = ID2D1Factory_CreatePathGeometry(factory, &path_geometry);
-    if (SUCCEEDED(hr))
-        hr = ID2D1PathGeometry_Open(path_geometry, &geometry_sink);
-    if (SUCCEEDED(hr))
-    {
-        hr = ID2D1Geometry_Simplify(geometry, D2D1_GEOMETRY_SIMPLIFICATION_OPTION_LINES,
-                transform, tolerance, (ID2D1SimplifiedGeometrySink *)geometry_sink);
-    }
-    if (SUCCEEDED(hr))
-        hr = ID2D1GeometrySink_Close(geometry_sink);
-    if (geometry_sink)
-        ID2D1GeometrySink_Release(geometry_sink);
-
-    if (SUCCEEDED(hr))
-    {
-        *ret = path_geometry;
-        ID2D1PathGeometry_AddRef(*ret);
-    }
-
-    if (path_geometry)
-        ID2D1PathGeometry_Release(path_geometry);
-    ID2D1Factory_Release(factory);
-
-    return hr;
-}
-
-static HRESULT d2d_geometry_tessellate(ID2D1Geometry *geometry, const D2D1_MATRIX_3X2_F *transform,
-        float tolerance, ID2D1TessellationSink *sink)
-{
-    ID2D1PathGeometry *path_geometry;
-    HRESULT hr;
-
-    if (SUCCEEDED(hr = d2d_geometry_get_simplified(geometry, transform, tolerance, &path_geometry)))
-    {
-        struct d2d_geometry *path_impl = unsafe_impl_from_ID2D1Geometry((ID2D1Geometry *)path_geometry);
-        D2D1_TRIANGLE t;
-
-        for (size_t i = 0; i < path_impl->fill.face_count; ++i)
-        {
-            const struct d2d_face *face = &path_impl->fill.faces[i];
-
-            t.point1 = path_impl->fill.vertices[face->v[0]];
-            t.point2 = path_impl->fill.vertices[face->v[1]];
-            t.point3 = path_impl->fill.vertices[face->v[2]];
-            ID2D1TessellationSink_AddTriangles(sink, &t, 1);
-        }
-
-        ID2D1PathGeometry_Release(path_geometry);
-    }
-
-    return hr;
-}
-
-static float d2d_triangle_area(const D2D1_TRIANGLE *triangle)
-{
-    D2D1_POINT_2F point2, point3;
-
-    /* Translate one vertex to origin */
-    point2.x = triangle->point2.x - triangle->point1.x;
-    point2.y = triangle->point2.y - triangle->point1.y;
-    point3.x = triangle->point3.x - triangle->point1.x;
-    point3.y = triangle->point3.y - triangle->point1.y;
-
-    return 0.5f * fabsf(point2.x * point3.y - point3.x * point2.y);
-}
-
-static HRESULT d2d_geometry_compute_area(ID2D1Geometry *geometry, const D2D1_MATRIX_3X2_F *transform,
-        float tolerance, float *ret)
-{
-    ID2D1PathGeometry *path_geometry;
-    float area = 0.0f;
-    HRESULT hr;
-
-    if (SUCCEEDED(hr = d2d_geometry_get_simplified(geometry, transform, tolerance, &path_geometry)))
-    {
-        struct d2d_geometry *path_impl = unsafe_impl_from_ID2D1Geometry((ID2D1Geometry *)path_geometry);
-        D2D1_TRIANGLE t;
-
-        for (size_t i = 0; i < path_impl->fill.face_count; ++i)
-        {
-            const struct d2d_face *face = &path_impl->fill.faces[i];
-
-            t.point1 = path_impl->fill.vertices[face->v[0]];
-            t.point2 = path_impl->fill.vertices[face->v[1]];
-            t.point3 = path_impl->fill.vertices[face->v[2]];
-            area += d2d_triangle_area(&t);
-        }
-
-        *ret = area;
-
-        ID2D1PathGeometry_Release(path_geometry);
-    }
-
-    return hr;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_path_geometry_Tessellate(ID2D1PathGeometry1 *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, ID2D1TessellationSink *sink)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1PathGeometry1(iface);
-
-    TRACE("iface %p, transform %p, tolerance %.8e, sink %p.\n", iface, transform, tolerance, sink);
-
-    return d2d_geometry_tessellate(&geometry->ID2D1Geometry_iface, transform, tolerance, sink);
+    FIXME("iface %p, transform %p, tolerance %.8e, sink %p stub!\n", iface, transform, tolerance, sink);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_path_geometry_CombineWithGeometry(ID2D1PathGeometry1 *iface,
@@ -3989,11 +3842,9 @@
 static HRESULT STDMETHODCALLTYPE d2d_path_geometry_ComputeArea(ID2D1PathGeometry1 *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, float *area)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1PathGeometry1(iface);
-
-    TRACE("iface %p, transform %p, tolerance %.8e, area %p.\n", iface, transform, tolerance, area);
-
-    return d2d_geometry_compute_area(&geometry->ID2D1Geometry_iface, transform, tolerance, area);
+    FIXME("iface %p, transform %p, tolerance %.8e, area %p stub!\n", iface, transform, tolerance, area);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_path_geometry_ComputeLength(ID2D1PathGeometry1 *iface,
@@ -4228,88 +4079,22 @@
     return E_NOTIMPL;
 }
 
-static void d2d_ellipse_to_segments(const D2D1_ELLIPSE *ellipse, D2D1_POINT_2F *start_point,
-        D2D1_BEZIER_SEGMENT *segments)
-{
-    const float coeff = 4.0f * (M_SQRT2 - 1.0f) / 3.0f;
-    D2D1_MATRIX_3X2_F m;
-    unsigned int i;
-
-    /* Use four Bézier segments to approximate a unit circle.
-       Endpoints tangents are tangential to the circle. Endpoints and the midpoint
-       are lying on the circle. */
-
-    d2d_point_set(start_point, -1.0f, 0.0f);
-
-    d2d_point_set(&segments[0].point1, -1.0f, -coeff);
-    d2d_point_set(&segments[0].point2, -coeff, -1.0f);
-    d2d_point_set(&segments[0].point3, 0.0f, -1.0f);
-
-    d2d_point_set(&segments[1].point1, coeff, -1.0f);
-    d2d_point_set(&segments[1].point2, 1.0f, -coeff);
-    d2d_point_set(&segments[1].point3, 1.0f, 0.0f);
-
-    d2d_point_set(&segments[2].point1, 1.0f, coeff);
-    d2d_point_set(&segments[2].point2, coeff, 1.0f);
-    d2d_point_set(&segments[2].point3, 0.0f, 1.0f);
-
-    d2d_point_set(&segments[3].point1, -coeff, 1.0f);
-    d2d_point_set(&segments[3].point2, -1.0f, coeff);
-    d2d_point_set(&segments[3].point3, start_point->x, start_point->y);
-
-    m._11 = ellipse->radiusX;
-    m._12 = 0.0f;
-    m._21 = 0.0f;
-    m._22 = ellipse->radiusY;
-    m._31 = ellipse->point.x;
-    m._32 = ellipse->point.y;
-
-    d2d_point_transform(start_point, &m, start_point->x, start_point->y);
-    for (i = 0; i < 4; ++i)
-    {
-        d2d_point_transform(&segments[i].point1, &m, segments[i].point1.x, segments[i].point1.y);
-        d2d_point_transform(&segments[i].point2, &m, segments[i].point2.x, segments[i].point2.y);
-        d2d_point_transform(&segments[i].point3, &m, segments[i].point3.x, segments[i].point3.y);
-    }
-}
-
 static HRESULT STDMETHODCALLTYPE d2d_ellipse_geometry_Simplify(ID2D1EllipseGeometry *iface,
         D2D1_GEOMETRY_SIMPLIFICATION_OPTION option, const D2D1_MATRIX_3X2_F *transform, float tolerance,
         ID2D1SimplifiedGeometrySink *sink)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1EllipseGeometry(iface);
-    struct d2d_figure figure = { 0 };
-    D2D1_BEZIER_SEGMENT segments[4];
-    D2D1_POINT_2F start_point;
-
-    TRACE("iface %p, option %#x, transform %p, tolerance %.8e, sink %p.\n",
+    FIXME("iface %p, option %#x, transform %p, tolerance %.8e, sink %p stub!\n",
             iface, option, transform, tolerance, sink);
 
-    d2d_ellipse_to_segments(&geometry->u.ellipse.ellipse, &start_point, segments);
-
-    if (!d2d_figure_begin(&figure, start_point, D2D1_FIGURE_BEGIN_FILLED))
-        return E_OUTOFMEMORY;
-    if (!d2d_figure_add_beziers(&figure, segments, ARRAY_SIZE(segments)))
-    {
-        d2d_figure_cleanup(&figure);
-        return E_OUTOFMEMORY;
-    }
-    d2d_figure_end(&figure, D2D1_FIGURE_END_CLOSED);
-
-    d2d_figure_simplify(&figure, option, transform, tolerance, sink);
-    d2d_figure_cleanup(&figure);
-
-    return S_OK;
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_ellipse_geometry_Tessellate(ID2D1EllipseGeometry *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, ID2D1TessellationSink *sink)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1EllipseGeometry(iface);
-
-    TRACE("iface %p, transform %p, tolerance %.8e, sink %p.\n", iface, transform, tolerance, sink);
-
-    return d2d_geometry_tessellate(&geometry->ID2D1Geometry_iface, transform, tolerance, sink);
+    FIXME("iface %p, transform %p, tolerance %.8e, sink %p stub!\n", iface, transform, tolerance, sink);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_ellipse_geometry_CombineWithGeometry(ID2D1EllipseGeometry *iface,
@@ -4333,11 +4118,9 @@
 static HRESULT STDMETHODCALLTYPE d2d_ellipse_geometry_ComputeArea(ID2D1EllipseGeometry *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, float *area)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1EllipseGeometry(iface);
-
-    TRACE("iface %p, transform %p, tolerance %.8e, area %p.\n", iface, transform, tolerance, area);
-
-    return d2d_geometry_compute_area(&geometry->ID2D1Geometry_iface, transform, tolerance, area);
+    FIXME("iface %p, transform %p, tolerance %.8e, area %p stub!\n", iface, transform, tolerance, area);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_ellipse_geometry_ComputeLength(ID2D1EllipseGeometry *iface,
@@ -4715,11 +4498,9 @@
 static HRESULT STDMETHODCALLTYPE d2d_rectangle_geometry_Tessellate(ID2D1RectangleGeometry *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, ID2D1TessellationSink *sink)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1RectangleGeometry(iface);
-
-    TRACE("iface %p, transform %p, tolerance %.8e, sink %p.\n", iface, transform, tolerance, sink);
-
-    return d2d_geometry_tessellate(&geometry->ID2D1Geometry_iface, transform, tolerance, sink);
+    FIXME("iface %p, transform %p, tolerance %.8e, sink %p stub!\n", iface, transform, tolerance, sink);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_rectangle_geometry_CombineWithGeometry(ID2D1RectangleGeometry *iface,
@@ -4743,30 +4524,9 @@
 static HRESULT STDMETHODCALLTYPE d2d_rectangle_geometry_ComputeArea(ID2D1RectangleGeometry *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, float *area)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1RectangleGeometry(iface);
-    const D2D_RECT_F *rect = &geometry->u.rectangle.rect;
-    D2D1_TRIANGLE triangle;
-    D2D1_MATRIX_3X2_F m;
-
-    TRACE("iface %p, transform %p, tolerance %.8e, area %p.\n", iface, transform, tolerance, area);
-
-    if (transform)
-    {
-        m = *transform;
-        m._31 = m._32 = 0.0f;
-
-        d2d_point_transform(&triangle.point1, &m, rect->left, rect->bottom);
-        d2d_point_transform(&triangle.point2, &m, rect->left, rect->top);
-        d2d_point_transform(&triangle.point3, &m, rect->right, rect->top);
-
-        *area = 2 * d2d_triangle_area(&triangle);
-    }
-    else
-    {
-        *area = fabsf((rect->right - rect->left) * (rect->bottom - rect->top));
-    }
-
-    return S_OK;
+    FIXME("iface %p, transform %p, tolerance %.8e, area %p stub!\n", iface, transform, tolerance, area);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_rectangle_geometry_ComputeLength(ID2D1RectangleGeometry *iface,
@@ -5009,83 +4769,22 @@
     return E_NOTIMPL;
 }
 
-static inline void d2d_point_translate(D2D1_POINT_2F *point, float x, float y)
-{
-    point->x += x;
-    point->y += y;
-}
-
-static inline void d2d_bezier_segment_translate(D2D1_BEZIER_SEGMENT *segment, float x, float y)
-{
-    d2d_point_translate(&segment->point1, x, y);
-    d2d_point_translate(&segment->point2, x, y);
-    d2d_point_translate(&segment->point3, x, y);
-}
-
 static HRESULT STDMETHODCALLTYPE d2d_rounded_rectangle_geometry_Simplify(ID2D1RoundedRectangleGeometry *iface,
         D2D1_GEOMETRY_SIMPLIFICATION_OPTION option, const D2D1_MATRIX_3X2_F *transform, float tolerance,
         ID2D1SimplifiedGeometrySink *sink)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1RoundedRectangleGeometry(iface);
-    const D2D1_ROUNDED_RECT *r = &geometry->u.rounded_rectangle.rounded_rect;
-    struct d2d_figure figure = { 0 };
-    D2D1_BEZIER_SEGMENT segments[4];
-    D2D1_POINT_2F start_point, p;
-    D2D1_ELLIPSE ellipse;
-    bool ret;
-
-    TRACE("iface %p, option %#x, transform %p, tolerance %.8e, sink %p.\n",
+    FIXME("iface %p, option %#x, transform %p, tolerance %.8e, sink %p stub!\n",
             iface, option, transform, tolerance, sink);
 
-    d2d_point_set(&ellipse.point, 0.0f, 0.0f);
-    ellipse.radiusX = r->radiusX;
-    ellipse.radiusY = r->radiusY;
-
-    d2d_ellipse_to_segments(&ellipse, &start_point, segments);
-
-    d2d_point_set(&p, r->rect.left + r->radiusX, r->rect.bottom + r->radiusY);
-    d2d_point_translate(&start_point, p.x, p.y);
-    d2d_bezier_segment_translate(&segments[0], p.x, p.y);
-    d2d_point_set(&p, r->rect.right - r->radiusX, r->rect.bottom + r->radiusY);
-    d2d_bezier_segment_translate(&segments[1], p.x, p.y);
-    d2d_point_set(&p, r->rect.right - r->radiusX, r->rect.top - r->radiusY);
-    d2d_bezier_segment_translate(&segments[2], p.x, p.y);
-    d2d_point_set(&p, r->rect.left + r->radiusX, r->rect.top - r->radiusY);
-    d2d_bezier_segment_translate(&segments[3], p.x, p.y);
-
-    ret = d2d_figure_begin(&figure, start_point, D2D1_FIGURE_BEGIN_FILLED);
-    ret = ret && d2d_figure_add_beziers(&figure, &segments[0], 1);
-    d2d_point_set(&p, r->rect.right - r->radiusX, r->rect.bottom);
-    ret = ret && d2d_figure_add_lines(&figure, &p, 1);
-    ret = ret && d2d_figure_add_beziers(&figure, &segments[1], 1);
-    d2d_point_set(&p, r->rect.right, r->rect.top - r->radiusY);
-    ret = ret && d2d_figure_add_lines(&figure, &p, 1);
-    ret = ret && d2d_figure_add_beziers(&figure, &segments[2], 1);
-    d2d_point_set(&p, r->rect.left + r->radiusX, r->rect.top);
-    ret = ret && d2d_figure_add_lines(&figure, &p, 1);
-    ret = ret && d2d_figure_add_beziers(&figure, &segments[3], 1);
-    if (!ret)
-    {
-        d2d_figure_cleanup(&figure);
-        return E_OUTOFMEMORY;
-    }
-
-    d2d_figure_end(&figure, D2D1_FIGURE_END_CLOSED);
-
-    d2d_figure_simplify(&figure, option, transform, tolerance, sink);
-    d2d_figure_cleanup(&figure);
-
-    return S_OK;
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_rounded_rectangle_geometry_Tessellate(ID2D1RoundedRectangleGeometry *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, ID2D1TessellationSink *sink)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1RoundedRectangleGeometry(iface);
-
-    TRACE("iface %p, transform %p, tolerance %.8e, sink %p.\n", iface, transform, tolerance, sink);
-
-    return d2d_geometry_tessellate(&geometry->ID2D1Geometry_iface, transform, tolerance, sink);
+    FIXME("iface %p, transform %p, tolerance %.8e, sink %p stub!\n", iface, transform, tolerance, sink);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_rounded_rectangle_geometry_CombineWithGeometry(
@@ -5109,11 +4808,9 @@
 static HRESULT STDMETHODCALLTYPE d2d_rounded_rectangle_geometry_ComputeArea(ID2D1RoundedRectangleGeometry *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, float *area)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1RoundedRectangleGeometry(iface);
-
-    TRACE("iface %p, transform %p, tolerance %.8e, area %p.\n", iface, transform, tolerance, area);
-
-    return d2d_geometry_compute_area(&geometry->ID2D1Geometry_iface, transform, tolerance, area);
+    FIXME("iface %p, transform %p, tolerance %.8e, area %p stub!\n", iface, transform, tolerance, area);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_rounded_rectangle_geometry_ComputeLength(ID2D1RoundedRectangleGeometry *iface,
@@ -5424,16 +5121,9 @@
 static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_Tessellate(ID2D1TransformedGeometry *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, ID2D1TessellationSink *sink)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1TransformedGeometry(iface);
-    D2D1_MATRIX_3X2_F g;
-
-    TRACE("iface %p, transform %p, tolerance %.8e, sink %p.\n", iface, transform, tolerance, sink);
-
-    g = geometry->transform;
-    if (transform)
-        d2d_matrix_multiply(&g, transform);
-
-    return ID2D1Geometry_Tessellate(geometry->u.transformed.src_geometry, transform, tolerance, sink);
+    FIXME("iface %p, transform %p, tolerance %.8e, sink %p stub!\n", iface, transform, tolerance, sink);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_CombineWithGeometry(ID2D1TransformedGeometry *iface,
@@ -5457,16 +5147,9 @@
 static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_ComputeArea(ID2D1TransformedGeometry *iface,
         const D2D1_MATRIX_3X2_F *transform, float tolerance, float *area)
 {
-    struct d2d_geometry *geometry = impl_from_ID2D1TransformedGeometry(iface);
-    D2D1_MATRIX_3X2_F g;
-
-    TRACE("iface %p, transform %p, tolerance %.8e, area %p.\n", iface, transform, tolerance, area);
-
-    g = geometry->transform;
-    if (transform)
-        d2d_matrix_multiply(&g, transform);
-
-    return ID2D1Geometry_ComputeArea(geometry->u.transformed.src_geometry, &g, tolerance, area);
+    FIXME("iface %p, transform %p, tolerance %.8e, area %p stub!\n", iface, transform, tolerance, area);
+
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d2d_transformed_geometry_ComputeLength(ID2D1TransformedGeometry *iface,
@@ -5846,95 +5529,3 @@
             || iface->lpVtbl == (const ID2D1GeometryVtbl *)&d2d_geometry_group_vtbl);
     return CONTAINING_RECORD(iface, struct d2d_geometry, ID2D1Geometry_iface);
 }
-
-static inline struct d2d_geometry_realization *impl_from_ID2D1GeometryRealization(
-        ID2D1GeometryRealization *iface)
-{
-    return CONTAINING_RECORD(iface, struct d2d_geometry_realization, ID2D1GeometryRealization_iface);
-}
-
-static HRESULT STDMETHODCALLTYPE d2d_geometry_realization_QueryInterface(
-        ID2D1GeometryRealization *iface, REFIID iid, void **out)
-{
-    TRACE("iface %p, iid %s, out %p.\n", iface, debugstr_guid(iid), out);
-
-    if (IsEqualGUID(iid, &IID_ID2D1GeometryRealization)
-            || IsEqualGUID(iid, &IID_ID2D1Resource)
-            || IsEqualGUID(iid, &IID_IUnknown))
-    {
-        ID2D1GeometryRealization_AddRef(iface);
-        *out = iface;
-        return S_OK;
-    }
-
-    WARN("%s not implemented, returning E_NOINTERFACE.\n", debugstr_guid(iid));
-
-    *out = NULL;
-    return E_NOINTERFACE;
-}
-
-static ULONG STDMETHODCALLTYPE d2d_geometry_realization_AddRef(ID2D1GeometryRealization *iface)
-{
-    struct d2d_geometry_realization *realization = impl_from_ID2D1GeometryRealization(iface);
-    ULONG refcount = InterlockedIncrement(&realization->refcount);
-
-    TRACE("%p increasing refcount to %lu.\n", iface, refcount);
-
-    return refcount;
-}
-
-static ULONG STDMETHODCALLTYPE d2d_geometry_realization_Release(ID2D1GeometryRealization *iface)
-{
-    struct d2d_geometry_realization *realization = impl_from_ID2D1GeometryRealization(iface);
-    ULONG refcount = InterlockedDecrement(&realization->refcount);
-
-    TRACE("%p decreasing refcount to %lu.\n", iface, refcount);
-
-    if (!refcount)
-    {
-        if (realization->stroke_style)
-            ID2D1StrokeStyle_Release(realization->stroke_style);
-        ID2D1Geometry_Release(realization->geometry);
-        ID2D1Factory_Release(realization->factory);
-        free(realization);
-    }
-
-    return refcount;
-}
-
-static void STDMETHODCALLTYPE d2d_geometry_realization_GetFactory(ID2D1GeometryRealization *iface,
-        ID2D1Factory **factory)
-{
-    struct d2d_geometry_realization *realization = impl_from_ID2D1GeometryRealization(iface);
-
-    TRACE("iface %p, factory %p.\n", iface, factory);
-
-    ID2D1Factory_AddRef(*factory = realization->factory);
-}
-
-static const ID2D1GeometryRealizationVtbl d2d_geometry_realization_vtbl =
-{
-    d2d_geometry_realization_QueryInterface,
-    d2d_geometry_realization_AddRef,
-    d2d_geometry_realization_Release,
-    d2d_geometry_realization_GetFactory,
-};
-
-HRESULT d2d_geometry_realization_init(struct d2d_geometry_realization *realization,
-        ID2D1Factory *factory, ID2D1Geometry *geometry)
-{
-    realization->ID2D1GeometryRealization_iface.lpVtbl = &d2d_geometry_realization_vtbl;
-    realization->refcount = 1;
-    ID2D1Factory_AddRef(realization->factory = factory);
-    ID2D1Geometry_AddRef(realization->geometry = geometry);
-
-    return S_OK;
-}
-
-struct d2d_geometry_realization *unsafe_impl_from_ID2D1GeometryRealization(ID2D1GeometryRealization *iface)
-{
-    if (!iface)
-        return NULL;
-    assert(iface->lpVtbl == &d2d_geometry_realization_vtbl);
-    return CONTAINING_RECORD(iface, struct d2d_geometry_realization, ID2D1GeometryRealization_iface);
-}
