diff --git a/dlls/mshtml/htmlelem.c b/dlls/mshtml/htmlelem.c
--- a/dlls/mshtml/htmlelem.c
+++ b/dlls/mshtml/htmlelem.c
@@ -301,7 +301,6 @@
 
     nsAString_InitDepend(&name_str, name);
     nsres = nsIDOMElement_HasAttribute(element->dom_element, &name_str, &r);
-    nsAString_Finish(&name_str);
     return variant_bool(NS_SUCCEEDED(nsres) && r);
 }
 
@@ -525,24 +524,6 @@
     return hres;
 }
 
-static HTMLDOMAttribute *find_attr_in_list(HTMLAttributeCollection *attrs, DISPID dispid, LONG *pos)
-{
-    HTMLDOMAttribute *iter, *ret = NULL;
-    struct list *list = &attrs->attrs;
-    unsigned i = 0;
-
-    LIST_FOR_EACH_ENTRY(iter, list, HTMLDOMAttribute, entry) {
-        if(iter->dispid == dispid) {
-            ret = iter;
-            break;
-        }
-        i++;
-    }
-    if(pos)
-        *pos = i;
-    return ret;
-}
-
 typedef struct {
     DispatchEx dispex;
     IHTMLRect IHTMLRect_iface;
@@ -550,8 +531,6 @@
 
     nsIDOMClientRect *nsrect;
 } HTMLRect;
-
-static HRESULT HTMLRectCollection_collection_item(DispatchEx*,ULONG,IDispatch**);
 
 static inline HTMLRect *impl_from_IHTMLRect(IHTMLRect *iface)
 {
@@ -783,7 +762,7 @@
     0
 };
 dispex_static_data_t ClientRect_dispex = {
-    .id         = OBJID_ClientRect,
+    .id         = PROT_ClientRect,
     .vtbl       = &ClientRect_dispex_vtbl,
     .disp_tid   = IHTMLRect_tid,
     .iface_tids = ClientRect_iface_tids,
@@ -817,6 +796,139 @@
     nsIDOMClientRectList *rect_list;
 } HTMLRectCollection;
 
+typedef struct {
+    IEnumVARIANT IEnumVARIANT_iface;
+
+    LONG ref;
+
+    ULONG iter;
+    HTMLRectCollection *col;
+} HTMLRectCollectionEnum;
+
+static inline HTMLRectCollectionEnum *HTMLRectCollectionEnum_from_IEnumVARIANT(IEnumVARIANT *iface)
+{
+    return CONTAINING_RECORD(iface, HTMLRectCollectionEnum, IEnumVARIANT_iface);
+}
+
+static HRESULT WINAPI HTMLRectCollectionEnum_QueryInterface(IEnumVARIANT *iface, REFIID riid, void **ppv)
+{
+    HTMLRectCollectionEnum *This = HTMLRectCollectionEnum_from_IEnumVARIANT(iface);
+
+    TRACE("(%p)->(%s %p)\n", This, debugstr_mshtml_guid(riid), ppv);
+
+    if(IsEqualGUID(riid, &IID_IUnknown)) {
+        *ppv = &This->IEnumVARIANT_iface;
+    }else if(IsEqualGUID(riid, &IID_IEnumVARIANT)) {
+        *ppv = &This->IEnumVARIANT_iface;
+    }else {
+        FIXME("(%p)->(%s %p)\n", This, debugstr_mshtml_guid(riid), ppv);
+        *ppv = NULL;
+        return E_NOINTERFACE;
+    }
+
+    IUnknown_AddRef((IUnknown*)*ppv);
+    return S_OK;
+}
+
+static ULONG WINAPI HTMLRectCollectionEnum_AddRef(IEnumVARIANT *iface)
+{
+    HTMLRectCollectionEnum *This = HTMLRectCollectionEnum_from_IEnumVARIANT(iface);
+    LONG ref = InterlockedIncrement(&This->ref);
+
+    TRACE("(%p) ref=%ld\n", This, ref);
+
+    return ref;
+}
+
+static ULONG WINAPI HTMLRectCollectionEnum_Release(IEnumVARIANT *iface)
+{
+    HTMLRectCollectionEnum *This = HTMLRectCollectionEnum_from_IEnumVARIANT(iface);
+    LONG ref = InterlockedDecrement(&This->ref);
+
+    TRACE("(%p) ref=%ld\n", This, ref);
+
+    if(!ref) {
+        IHTMLRectCollection_Release(&This->col->IHTMLRectCollection_iface);
+        free(This);
+    }
+
+    return ref;
+}
+
+static HRESULT WINAPI HTMLRectCollectionEnum_Next(IEnumVARIANT *iface, ULONG celt, VARIANT *rgVar, ULONG *pCeltFetched)
+{
+    HTMLRectCollectionEnum *This = HTMLRectCollectionEnum_from_IEnumVARIANT(iface);
+    VARIANT index;
+    HRESULT hres;
+    ULONG num, i;
+    UINT32 len;
+
+    TRACE("(%p)->(%lu %p %p)\n", This, celt, rgVar, pCeltFetched);
+
+    nsIDOMClientRectList_GetLength(This->col->rect_list, &len);
+    num = min(len - This->iter, celt);
+    V_VT(&index) = VT_I4;
+
+    for(i = 0; i < num; i++) {
+        V_I4(&index) = This->iter + i;
+        hres = IHTMLRectCollection_item(&This->col->IHTMLRectCollection_iface, &index, &rgVar[i]);
+        if(FAILED(hres)) {
+            while(i--)
+                VariantClear(&rgVar[i]);
+            return hres;
+        }
+    }
+
+    This->iter += num;
+    if(pCeltFetched)
+        *pCeltFetched = num;
+    return num == celt ? S_OK : S_FALSE;
+}
+
+static HRESULT WINAPI HTMLRectCollectionEnum_Skip(IEnumVARIANT *iface, ULONG celt)
+{
+    HTMLRectCollectionEnum *This = HTMLRectCollectionEnum_from_IEnumVARIANT(iface);
+    UINT32 len;
+
+    TRACE("(%p)->(%lu)\n", This, celt);
+
+    nsIDOMClientRectList_GetLength(This->col->rect_list, &len);
+    if(This->iter + celt > len) {
+        This->iter = len;
+        return S_FALSE;
+    }
+
+    This->iter += celt;
+    return S_OK;
+}
+
+static HRESULT WINAPI HTMLRectCollectionEnum_Reset(IEnumVARIANT *iface)
+{
+    HTMLRectCollectionEnum *This = HTMLRectCollectionEnum_from_IEnumVARIANT(iface);
+
+    TRACE("(%p)->()\n", This);
+
+    This->iter = 0;
+    return S_OK;
+}
+
+static HRESULT WINAPI HTMLRectCollectionEnum_Clone(IEnumVARIANT *iface, IEnumVARIANT **ppEnum)
+{
+    HTMLRectCollectionEnum *This = HTMLRectCollectionEnum_from_IEnumVARIANT(iface);
+    FIXME("(%p)->(%p)\n", This, ppEnum);
+    return E_NOTIMPL;
+}
+
+static const IEnumVARIANTVtbl HTMLRectCollectionEnumVtbl = {
+    HTMLRectCollectionEnum_QueryInterface,
+    HTMLRectCollectionEnum_AddRef,
+    HTMLRectCollectionEnum_Release,
+    HTMLRectCollectionEnum_Next,
+    HTMLRectCollectionEnum_Skip,
+    HTMLRectCollectionEnum_Reset,
+    HTMLRectCollectionEnum_Clone
+};
+
 static inline HTMLRectCollection *impl_from_IHTMLRectCollection(IHTMLRectCollection *iface)
 {
     return CONTAINING_RECORD(iface, HTMLRectCollection, IHTMLRectCollection_iface);
@@ -841,15 +953,32 @@
 static HRESULT WINAPI HTMLRectCollection_get__newEnum(IHTMLRectCollection *iface, IUnknown **p)
 {
     HTMLRectCollection *This = impl_from_IHTMLRectCollection(iface);
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    return create_enum_variant(&This->dispex, p);
+    HTMLRectCollectionEnum *ret;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    ret = malloc(sizeof(*ret));
+    if(!ret)
+        return E_OUTOFMEMORY;
+
+    ret->IEnumVARIANT_iface.lpVtbl = &HTMLRectCollectionEnumVtbl;
+    ret->ref = 1;
+    ret->iter = 0;
+
+    HTMLRectCollection_AddRef(&This->IHTMLRectCollection_iface);
+    ret->col = This;
+
+    *p = (IUnknown*)&ret->IEnumVARIANT_iface;
+    return S_OK;
 }
 
 static HRESULT WINAPI HTMLRectCollection_item(IHTMLRectCollection *iface, VARIANT *index, VARIANT *result)
 {
     HTMLRectCollection *This = impl_from_IHTMLRectCollection(iface);
+    nsIDOMClientRect *nsrect;
+    IHTMLRect *rect;
+    nsresult nsres;
+    HRESULT hres;
 
     TRACE("(%p)->(%s %p)\n", This, debugstr_variant(index), result);
 
@@ -858,8 +987,22 @@
         return E_NOTIMPL;
     }
 
+    nsres = nsIDOMClientRectList_Item(This->rect_list, V_I4(index), &nsrect);
+    if(NS_FAILED(nsres))
+        return map_nsresult(nsres);
+    if(!nsrect) {
+        V_VT(result) = VT_NULL;
+        return S_OK;
+    }
+
+    hres = create_html_rect(nsrect, &This->dispex, &rect);
+    nsIDOMClientRect_Release(nsrect);
+    if(FAILED(hres))
+        return hres;
+
     V_VT(result) = VT_DISPATCH;
-    return HTMLRectCollection_collection_item(&This->dispex, V_I4(index), &V_DISPATCH(result));
+    V_DISPATCH(result) = (IDispatch *)rect;
+    return S_OK;
 }
 
 static const IHTMLRectCollectionVtbl HTMLRectCollectionVtbl = {
@@ -938,46 +1081,34 @@
     TRACE("(%p)->(%lx %lx %x %p %p %p %p)\n", This, id, lcid, flags, params, res, ei, caller);
 
     switch(flags) {
-    case DISPATCH_PROPERTYGET:
+    case DISPATCH_PROPERTYGET: {
+        nsIDOMClientRect *rect;
+        IHTMLRect *html_rect;
+        nsresult nsres;
+        HRESULT hres;
+
+        nsres = nsIDOMClientRectList_Item(This->rect_list, id - MSHTML_DISPID_CUSTOM_MIN, &rect);
+        if(NS_FAILED(nsres) || !rect) {
+            WARN("Unknown item\n");
+            return DISP_E_MEMBERNOTFOUND;
+        }
+
+        hres = create_html_rect(rect, &This->dispex, &html_rect);
+        nsIDOMClientRect_Release(rect);
+        if(FAILED(hres))
+            return hres;
+
         V_VT(res) = VT_DISPATCH;
-        return HTMLRectCollection_collection_item(&This->dispex, id - MSHTML_DISPID_CUSTOM_MIN, &V_DISPATCH(res));
+        V_DISPATCH(res) = (IDispatch*)html_rect;
+        break;
+    }
+
     default:
         FIXME("unimplemented flags %x\n", flags);
         return E_NOTIMPL;
     }
 
     return S_OK;
-}
-
-static ULONG HTMLRectCollection_collection_len(DispatchEx *dispex)
-{
-    HTMLRectCollection *This = HTMLRectCollection_from_DispatchEx(dispex);
-    nsresult nsres;
-    UINT32 len;
-
-    nsres = nsIDOMClientRectList_GetLength(This->rect_list, &len);
-    assert(nsres == NS_OK);
-    return len;
-}
-
-static HRESULT HTMLRectCollection_collection_item(DispatchEx *dispex, ULONG index, IDispatch **p)
-{
-    HTMLRectCollection *This = HTMLRectCollection_from_DispatchEx(dispex);
-    nsIDOMClientRect *nsrect;
-    nsresult nsres;
-    HRESULT hres;
-
-    nsres = nsIDOMClientRectList_Item(This->rect_list, index, &nsrect);
-    if(NS_FAILED(nsres))
-        return map_nsresult(nsres);
-    if(!nsrect) {
-        *p = NULL;
-        return S_OK;
-    }
-
-    hres = create_html_rect(nsrect, &This->dispex, (IHTMLRect**)p);
-    nsIDOMClientRect_Release(nsrect);
-    return hres;
 }
 
 static const dispex_static_data_vtbl_t HTMLRectCollection_dispex_vtbl = {
@@ -988,15 +1119,13 @@
     .get_dispid       = HTMLRectCollection_get_dispid,
     .get_prop_desc    = dispex_index_prop_desc,
     .invoke           = HTMLRectCollection_invoke,
-    .collection_len   = HTMLRectCollection_collection_len,
-    .collection_item  = HTMLRectCollection_collection_item,
 };
 static const tid_t ClientRectList_iface_tids[] = {
     IHTMLRectCollection_tid,
     0
 };
 dispex_static_data_t ClientRectList_dispex = {
-    .id         = OBJID_ClientRectList,
+    .id         = PROT_ClientRectList,
     .vtbl       = &HTMLRectCollection_dispex_vtbl,
     .disp_tid   = IHTMLRectCollection_tid,
     .iface_tids = ClientRectList_iface_tids,
@@ -1184,9 +1313,6 @@
     TRACE("(%p)->(%s %lx %p)\n", This, debugstr_w(strAttributeName), lFlags, pfSuccess);
 
     if(compat_mode < COMPAT_MODE_IE9 || !This->dom_element) {
-        HTMLAttributeCollection *attrs;
-        HTMLDOMAttribute *attr;
-
         hres = dispex_get_id(&This->node.event_target.dispex, translate_attr_name(strAttributeName, compat_mode),
                              lFlags & ATTRFLAG_CASESENSITIVE ? fdexNameCaseSensitive : fdexNameCaseInsensitive, &id);
         if(hres == DISP_E_UNKNOWNNAME) {
@@ -1195,26 +1321,6 @@
         }
         if(FAILED(hres))
             return hres;
-
-        hres = HTMLElement_get_attr_col(&This->node, &attrs);
-        if(FAILED(hres))
-            return hres;
-        attr = find_attr_in_list(attrs, id, NULL);
-        IHTMLAttributeCollection_Release(&attrs->IHTMLAttributeCollection_iface);
-
-        if(attr) {
-            hres = get_elem_attr_value_by_dispid(This, id, &attr->value);
-            if(FAILED(hres))
-                return hres;
-            if(!attr->name) {
-                hres = dispex_prop_name(&This->node.event_target.dispex, id, &attr->name);
-                if(FAILED(hres))
-                    return hres;
-            }
-            list_remove(&attr->entry);
-            IHTMLDOMNode_Release(&attr->elem->node.IHTMLDOMNode_iface);
-            attr->elem = NULL;
-        }
 
         if(id == DISPID_IHTMLELEMENT_STYLE) {
             IHTMLStyle *style;
@@ -4284,7 +4390,7 @@
         IHTMLDOMAttribute **ppretAttribute)
 {
     HTMLElement *This = impl_from_IHTMLElement4(iface);
-    HTMLDOMAttribute *attr, *replace;
+    HTMLDOMAttribute *attr, *iter, *replace = NULL;
     HTMLAttributeCollection *attrs;
     DISPID dispid;
     HRESULT hres;
@@ -4295,31 +4401,6 @@
     if(!attr)
         return E_INVALIDARG;
 
-    if(attr->dom_attr) {
-        HTMLDOMAttribute *prev_attr;
-        nsIDOMAttr *prev_nsattr;
-        nsresult nsres;
-
-        if(!This->dom_element) {
-            FIXME("no DOM element\n");
-            return E_NOTIMPL;
-        }
-
-        nsres = nsIDOMElement_SetAttributeNode(This->dom_element, attr->dom_attr, &prev_nsattr);
-        if(NS_FAILED(nsres))
-            return map_nsresult(nsres);
-        if(!prev_nsattr) {
-            *ppretAttribute = NULL;
-            return S_OK;
-        }
-
-        hres = get_attr_node(prev_nsattr, &prev_attr);
-        nsIDOMAttr_Release(prev_nsattr);
-        if(SUCCEEDED(hres))
-            *ppretAttribute = &prev_attr->IHTMLDOMAttribute_iface;
-        return hres;
-    }
-
     if(attr->elem) {
         WARN("Tried to set already attached attribute.\n");
         return E_INVALIDARG;
@@ -4333,7 +4414,13 @@
     if(FAILED(hres))
         return hres;
 
-    replace = find_attr_in_list(attrs, dispid, NULL);
+    LIST_FOR_EACH_ENTRY(iter, &attrs->attrs, HTMLDOMAttribute, entry) {
+        if(iter->dispid == dispid) {
+            replace = iter;
+            break;
+        }
+    }
+
     if(replace) {
         hres = get_elem_attr_value_by_dispid(This, dispid, &replace->value);
         if(FAILED(hres)) {
@@ -4630,29 +4717,21 @@
 static HRESULT WINAPI HTMLElement6_getAttributeNode(IHTMLElement6 *iface, BSTR strAttributeName, IHTMLDOMAttribute2 **ppretAttribute)
 {
     HTMLElement *This = impl_from_IHTMLElement6(iface);
-    HTMLDOMAttribute *attr;
-    nsIDOMAttr *nsattr = NULL;
-    nsresult nsres;
+    IHTMLDOMAttribute *attr;
     HRESULT hres;
 
-    TRACE("(%p)->(%s %p)\n", This, debugstr_w(strAttributeName), ppretAttribute);
-
-    if(This->dom_element) {
-        nsAString nsstr;
-        nsAString_Init(&nsstr, strAttributeName);
-        nsres = nsIDOMElement_GetAttributeNode(This->dom_element, &nsstr, &nsattr);
-        nsAString_Finish(&nsstr);
-        if(NS_FAILED(nsres))
-            return map_nsresult(nsres);
-    }
-    if(!nsattr) {
+    WARN("(%p)->(%s %p) forwarding to IHTMLElement4\n", This, debugstr_w(strAttributeName), ppretAttribute);
+
+    hres = IHTMLElement4_getAttributeNode(&This->IHTMLElement4_iface, strAttributeName, &attr);
+    if(FAILED(hres))
+        return hres;
+
+    if(attr) {
+        hres = IHTMLDOMAttribute_QueryInterface(attr, &IID_IHTMLDOMAttribute2, (void**)ppretAttribute);
+        IHTMLDOMAttribute_Release(attr);
+    }else {
         *ppretAttribute = NULL;
-        return S_OK;
-    }
-    hres = get_attr_node(nsattr, &attr);
-    nsIDOMAttr_Release(nsattr);
-    if(SUCCEEDED(hres))
-        *ppretAttribute = &attr->IHTMLDOMAttribute2_iface;
+    }
     return hres;
 }
 
@@ -6588,14 +6667,6 @@
     return hres;
 }
 
-static const dispex_hook_t elem_traversal_noattr_hooks[] = {
-    {DISPID_IELEMENTTRAVERSAL_FIRSTELEMENTCHILD,      .noattr = TRUE},
-    {DISPID_IELEMENTTRAVERSAL_LASTELEMENTCHILD,       .noattr = TRUE},
-    {DISPID_IELEMENTTRAVERSAL_PREVIOUSELEMENTSIBLING, .noattr = TRUE},
-    {DISPID_IELEMENTTRAVERSAL_NEXTELEMENTSIBLING,     .noattr = TRUE},
-    {DISPID_IELEMENTTRAVERSAL_CHILDELEMENTCOUNT,      .noattr = TRUE}
-};
-
 void HTMLElement_init_dispex_info(dispex_data_t *info, compat_mode_t mode)
 {
     static const dispex_hook_t elem6_ie9_hooks[] = {
@@ -6628,33 +6699,10 @@
 
         /* IE9+ */
         {DISPID_IHTMLELEMENT_TOSTRING,     NULL},
-
-        /* Common for all modes */
-        {DISPID_IHTMLELEMENT_PARENTELEMENT,  .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_CLASSNAME,      .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_TAGNAME,        .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_OFFSETLEFT,     .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_OFFSETTOP,      .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_OFFSETWIDTH,    .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_OFFSETHEIGHT,   .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_OFFSETPARENT,   .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_DOCUMENT,       .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_SOURCEINDEX,    .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_RECORDNUMBER,   .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_INNERHTML,      .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_INNERTEXT,      .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_OUTERHTML,      .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_OUTERTEXT,      .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_PARENTTEXTEDIT, .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_ISTEXTEDIT,     .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_FILTERS,        .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_CHILDREN,       .noattr = TRUE},
-        {DISPID_IHTMLELEMENT_ALL,            .noattr = TRUE},
         {DISPID_UNKNOWN}
     };
     const dispex_hook_t *const elem_ie10_hooks = elem_ie11_hooks + 10;
     const dispex_hook_t *const elem_ie9_hooks  = elem_ie10_hooks + 2;
-    const dispex_hook_t *const elem_hooks      = elem_ie9_hooks  + 1;
     static const dispex_hook_t elem2_ie11_hooks[] = {
         {DISPID_IHTMLELEMENT2_ONLOSECAPTURE},
         {DISPID_IHTMLELEMENT2_ONPROPERTYCHANGE},
@@ -6682,28 +6730,10 @@
         {DISPID_IHTMLELEMENT2_SETEXPRESSION,    NULL},
         {DISPID_IHTMLELEMENT2_GETEXPRESSION,    NULL},
         {DISPID_IHTMLELEMENT2_REMOVEEXPRESSION, NULL},
-
-        /* Common for all modes */
-        {DISPID_IHTMLELEMENT2_CURRENTSTYLE,    .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_SCOPENAME,       .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_SCROLLHEIGHT,    .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_SCROLLWIDTH,     .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_SCROLLTOP,       .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_SCROLLLEFT,      .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_RUNTIMESTYLE,    .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_CANHAVECHILDREN, .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_BEHAVIORURNS,    .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_TAGURN,          .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_CLIENTHEIGHT,    .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_CLIENTWIDTH,     .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_CLIENTTOP,       .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_CLIENTLEFT,      .noattr = TRUE},
-        {DISPID_IHTMLELEMENT2_READYSTATE,      .noattr = TRUE},
         {DISPID_UNKNOWN}
     };
     const dispex_hook_t *const elem2_ie10_hooks = elem2_ie11_hooks + 15;
     const dispex_hook_t *const elem2_ie9_hooks  = elem2_ie10_hooks + 4;
-    const dispex_hook_t *const elem2_hooks      = elem2_ie9_hooks  + 3;
     static const dispex_hook_t elem3_ie11_hooks[] = {
         {DISPID_IHTMLELEMENT3_ONLAYOUTCOMPLETE},
         {DISPID_IHTMLELEMENT3_ONMOVE},
@@ -6716,17 +6746,9 @@
 
         /* IE9+ */
         {DISPID_IHTMLELEMENT3_ONPAGE},
-
-        /* Common for all modes */
-        {DISPID_IHTMLELEMENT3_ISMULTILINE,       .noattr = TRUE},
-        {DISPID_IHTMLELEMENT3_CANHAVEHTML,       .noattr = TRUE},
-        {DISPID_IHTMLELEMENT3_ISCONTENTEDITABLE, .noattr = TRUE},
-        {DISPID_IHTMLELEMENT3_ISDISABLED,        .noattr = TRUE},
-
         {DISPID_UNKNOWN}
     };
     const dispex_hook_t *const elem3_ie9_hooks = elem3_ie11_hooks + 8;
-    const dispex_hook_t *const elem3_hooks     = elem3_ie9_hooks  + 1;
     static const dispex_hook_t elem7_ie11_hooks[] = {
         {DISPID_IHTMLELEMENT7_ONMSPOINTERHOVER},
 
@@ -6741,22 +6763,18 @@
         {DISPID_UNKNOWN}
     };
     const dispex_hook_t *const elem7_ie10_hooks = elem7_ie11_hooks + 1;
-    static const dispex_hook_t unique_name_noattr_hooks[] = {
-        {DISPID_IHTMLUNIQUENAME_UNIQUENUMBER, .noattr = TRUE},
-        {DISPID_IHTMLUNIQUENAME_UNIQUEID,     .noattr = TRUE}
-    };
 
     HTMLDOMNode_init_dispex_info(info, mode);
 
     dispex_info_add_interface(info, IHTMLElement2_tid, mode >= COMPAT_MODE_IE11 ? elem2_ie11_hooks :
                                                        mode >= COMPAT_MODE_IE10 ? elem2_ie10_hooks :
-                                                       mode >= COMPAT_MODE_IE9  ? elem2_ie9_hooks  : elem2_hooks);
+                                                       mode >= COMPAT_MODE_IE9  ? elem2_ie9_hooks  : NULL);
     if(mode >= COMPAT_MODE_IE8)
         dispex_info_add_interface(info, IElementSelector_tid, NULL);
 
     if(mode >= COMPAT_MODE_IE9) {
         dispex_info_add_interface(info, IHTMLElement6_tid, mode >= COMPAT_MODE_IE10 ? elem6_ie10_hooks : elem6_ie9_hooks);
-        dispex_info_add_interface(info, IElementTraversal_tid, elem_traversal_noattr_hooks);
+        dispex_info_add_interface(info, IElementTraversal_tid, NULL);
     }
 
     if(mode >= COMPAT_MODE_IE10)
@@ -6767,13 +6785,13 @@
     }
 
     dispex_info_add_interface(info, IHTMLElement3_tid, mode >= COMPAT_MODE_IE11 ? elem3_ie11_hooks :
-                                                       mode >= COMPAT_MODE_IE9  ? elem3_ie9_hooks  : elem3_hooks);
+                                                       mode >= COMPAT_MODE_IE9  ? elem3_ie9_hooks  : NULL);
     dispex_info_add_interface(info, IHTMLElement_tid, mode >= COMPAT_MODE_IE11 ? elem_ie11_hooks :
                                                       mode >= COMPAT_MODE_IE10 ? elem_ie10_hooks :
-                                                      mode >= COMPAT_MODE_IE9  ? elem_ie9_hooks  : elem_hooks);
+                                                      mode >= COMPAT_MODE_IE9  ? elem_ie9_hooks  : NULL);
     dispex_info_add_interface(info, IHTMLElement4_tid, NULL);
     dispex_info_add_interface(info, IHTMLDOMNode_tid, NULL);
-    dispex_info_add_interface(info, IHTMLUniqueName_tid, unique_name_noattr_hooks);
+    dispex_info_add_interface(info, IHTMLUniqueName_tid, NULL);
 }
 
 static const event_target_vtbl_t HTMLElement_event_target_vtbl = {
@@ -7190,7 +7208,7 @@
     0
 };
 dispex_static_data_t DOMTokenList_dispex = {
-    .id              = OBJID_DOMTokenList,
+    .id              = PROT_DOMTokenList,
     .vtbl            = &token_list_dispex_vtbl,
     .disp_tid        = IWineDOMTokenList_tid,
     .iface_tids      = DOMTokenList_tids,
@@ -7392,19 +7410,19 @@
     dispex_info_add_dispids(info, IHTMLElement_tid, elem_dispids);
     dispex_info_add_dispids(info, IHTMLElement4_tid, elem4_dispids);
     dispex_info_add_interface(info, IElementSelector_tid, NULL);
-    dispex_info_add_interface(info, IElementTraversal_tid, elem_traversal_noattr_hooks);
+    dispex_info_add_interface(info, IElementTraversal_tid, NULL);
 }
 
 dispex_static_data_t Element_dispex = {
-    .id           = OBJID_Element,
-    .prototype_id = OBJID_Node,
+    .id           = PROT_Element,
+    .prototype_id = PROT_Node,
     .vtbl         = &HTMLElement_event_target_vtbl.dispex_vtbl,
     .init_info    = Element_init_dispex_info,
 };
 
 dispex_static_data_t HTMLElement_dispex = {
-    .id           = OBJID_HTMLElement,
-    .prototype_id = OBJID_Element,
+    .id           = PROT_HTMLElement,
+    .prototype_id = PROT_Element,
     .vtbl         = &HTMLElement_event_target_vtbl.dispex_vtbl,
     .disp_tid     = DispHTMLUnknownElement_tid,
     .init_info    = HTMLElement_init_dispex_info,
@@ -7681,20 +7699,9 @@
     return S_OK;
 }
 
-static inline BOOL is_valid_attr_dispid(HTMLAttributeCollection *col, DISPID id)
-{
-    if(get_dispid_type(id) != DISPEXPROP_BUILTIN)
-        return TRUE;
-
-    if(dispex_builtin_is_noattr(&col->elem->node.event_target.dispex, id))
-        return FALSE;
-
-    return TRUE;
-}
-
-static HRESULT get_attr_dispid_by_idx(HTMLAttributeCollection *This, LONG *idx, DISPID *dispid)
-{
-    DISPID id = DISPID_STARTENUM;
+static HRESULT get_attr_dispid_by_relative_idx(HTMLAttributeCollection *This, LONG *idx, DISPID start, DISPID *dispid)
+{
+    DISPID id = start;
     LONG len = -1;
     HRESULT hres;
 
@@ -7706,8 +7713,6 @@
             return hres;
         else if(hres == S_FALSE)
             break;
-        else if(!is_valid_attr_dispid(This, id))
-            continue;
 
         len++;
         if(len == *idx)
@@ -7721,6 +7726,11 @@
 
     *idx = len+1;
     return S_OK;
+}
+
+static HRESULT get_attr_dispid_by_idx(HTMLAttributeCollection *This, LONG *idx, DISPID *dispid)
+{
+    return get_attr_dispid_by_relative_idx(This, idx, DISPID_STARTENUM, dispid);
 }
 
 static inline HRESULT get_attr_dispid_by_name(HTMLAttributeCollection *This, const WCHAR *name, DISPID *id)
@@ -7739,23 +7749,189 @@
         }
     }
 
-    hres = dispex_get_id(&This->elem->node.event_target.dispex, name, fdexNameCaseInsensitive, id);
-    return (FAILED(hres) || is_valid_attr_dispid(This, *id)) ? hres : DISP_E_UNKNOWNNAME;
+    return dispex_get_id(&This->elem->node.event_target.dispex, name, fdexNameCaseInsensitive, id);
 }
 
 static inline HRESULT get_domattr(HTMLAttributeCollection *This, DISPID id, LONG *list_pos, HTMLDOMAttribute **attr)
 {
+    HTMLDOMAttribute *iter;
+    LONG pos = 0;
     HRESULT hres;
 
-    if(!(*attr = find_attr_in_list(This, id, list_pos))) {
+    *attr = NULL;
+    LIST_FOR_EACH_ENTRY(iter, &This->attrs, HTMLDOMAttribute, entry) {
+        if(iter->dispid == id) {
+            *attr = iter;
+            break;
+        }
+        pos++;
+    }
+
+    if(!*attr) {
         hres = HTMLDOMAttribute_Create(NULL, This->elem, id, This->elem->node.doc, attr);
         if(FAILED(hres))
             return hres;
     }
 
     IHTMLDOMAttribute_AddRef(&(*attr)->IHTMLDOMAttribute_iface);
-    return S_OK;
-}
+    if(list_pos)
+        *list_pos = pos;
+    return S_OK;
+}
+
+typedef struct {
+    IEnumVARIANT IEnumVARIANT_iface;
+
+    LONG ref;
+
+    ULONG iter;
+    DISPID iter_dispid;
+    HTMLAttributeCollection *col;
+} HTMLAttributeCollectionEnum;
+
+static inline HTMLAttributeCollectionEnum *HTMLAttributeCollectionEnum_from_IEnumVARIANT(IEnumVARIANT *iface)
+{
+    return CONTAINING_RECORD(iface, HTMLAttributeCollectionEnum, IEnumVARIANT_iface);
+}
+
+static HRESULT WINAPI HTMLAttributeCollectionEnum_QueryInterface(IEnumVARIANT *iface, REFIID riid, void **ppv)
+{
+    HTMLAttributeCollectionEnum *This = HTMLAttributeCollectionEnum_from_IEnumVARIANT(iface);
+
+    TRACE("(%p)->(%s %p)\n", This, debugstr_mshtml_guid(riid), ppv);
+
+    if(IsEqualGUID(riid, &IID_IUnknown)) {
+        *ppv = &This->IEnumVARIANT_iface;
+    }else if(IsEqualGUID(riid, &IID_IEnumVARIANT)) {
+        *ppv = &This->IEnumVARIANT_iface;
+    }else {
+        FIXME("(%p)->(%s %p)\n", This, debugstr_mshtml_guid(riid), ppv);
+        *ppv = NULL;
+        return E_NOINTERFACE;
+    }
+
+    IUnknown_AddRef((IUnknown*)*ppv);
+    return S_OK;
+}
+
+static ULONG WINAPI HTMLAttributeCollectionEnum_AddRef(IEnumVARIANT *iface)
+{
+    HTMLAttributeCollectionEnum *This = HTMLAttributeCollectionEnum_from_IEnumVARIANT(iface);
+    LONG ref = InterlockedIncrement(&This->ref);
+
+    TRACE("(%p) ref=%ld\n", This, ref);
+
+    return ref;
+}
+
+static ULONG WINAPI HTMLAttributeCollectionEnum_Release(IEnumVARIANT *iface)
+{
+    HTMLAttributeCollectionEnum *This = HTMLAttributeCollectionEnum_from_IEnumVARIANT(iface);
+    LONG ref = InterlockedDecrement(&This->ref);
+
+    TRACE("(%p) ref=%ld\n", This, ref);
+
+    if(!ref) {
+        IHTMLAttributeCollection_Release(&This->col->IHTMLAttributeCollection_iface);
+        free(This);
+    }
+
+    return ref;
+}
+
+static HRESULT WINAPI HTMLAttributeCollectionEnum_Next(IEnumVARIANT *iface, ULONG celt, VARIANT *rgVar, ULONG *pCeltFetched)
+{
+    HTMLAttributeCollectionEnum *This = HTMLAttributeCollectionEnum_from_IEnumVARIANT(iface);
+    DISPID tmp, dispid = This->iter_dispid;
+    HTMLDOMAttribute *attr;
+    LONG rel_index = 0;
+    HRESULT hres;
+    ULONG i;
+
+    TRACE("(%p)->(%lu %p %p)\n", This, celt, rgVar, pCeltFetched);
+
+    for(i = 0; i < celt; i++) {
+        hres = get_attr_dispid_by_relative_idx(This->col, &rel_index, dispid, &tmp);
+        if(SUCCEEDED(hres)) {
+            dispid = tmp;
+            hres = get_domattr(This->col, dispid, NULL, &attr);
+        }
+        else if(hres == DISP_E_UNKNOWNNAME)
+            break;
+
+        if(FAILED(hres)) {
+            while(i--)
+                VariantClear(&rgVar[i]);
+            return hres;
+        }
+
+        V_VT(&rgVar[i]) = VT_DISPATCH;
+        V_DISPATCH(&rgVar[i]) = (IDispatch*)&attr->IHTMLDOMAttribute_iface;
+    }
+
+    This->iter += i;
+    This->iter_dispid = dispid;
+    if(pCeltFetched)
+        *pCeltFetched = i;
+    return i == celt ? S_OK : S_FALSE;
+}
+
+static HRESULT WINAPI HTMLAttributeCollectionEnum_Skip(IEnumVARIANT *iface, ULONG celt)
+{
+    HTMLAttributeCollectionEnum *This = HTMLAttributeCollectionEnum_from_IEnumVARIANT(iface);
+    LONG remaining, rel_index;
+    DISPID dispid;
+    HRESULT hres;
+
+    TRACE("(%p)->(%lu)\n", This, celt);
+
+    if(!celt)
+        return S_OK;
+
+    rel_index = -1;
+    hres = get_attr_dispid_by_relative_idx(This->col, &rel_index, This->iter_dispid, NULL);
+    if(FAILED(hres))
+        return hres;
+    remaining = min(celt, rel_index);
+
+    if(remaining) {
+        rel_index = remaining - 1;
+        hres = get_attr_dispid_by_relative_idx(This->col, &rel_index, This->iter_dispid, &dispid);
+        if(FAILED(hres))
+            return hres;
+        This->iter += remaining;
+        This->iter_dispid = dispid;
+    }
+    return celt > remaining ? S_FALSE : S_OK;
+}
+
+static HRESULT WINAPI HTMLAttributeCollectionEnum_Reset(IEnumVARIANT *iface)
+{
+    HTMLAttributeCollectionEnum *This = HTMLAttributeCollectionEnum_from_IEnumVARIANT(iface);
+
+    TRACE("(%p)->()\n", This);
+
+    This->iter = 0;
+    This->iter_dispid = DISPID_STARTENUM;
+    return S_OK;
+}
+
+static HRESULT WINAPI HTMLAttributeCollectionEnum_Clone(IEnumVARIANT *iface, IEnumVARIANT **ppEnum)
+{
+    HTMLAttributeCollectionEnum *This = HTMLAttributeCollectionEnum_from_IEnumVARIANT(iface);
+    FIXME("(%p)->(%p)\n", This, ppEnum);
+    return E_NOTIMPL;
+}
+
+static const IEnumVARIANTVtbl HTMLAttributeCollectionEnumVtbl = {
+    HTMLAttributeCollectionEnum_QueryInterface,
+    HTMLAttributeCollectionEnum_AddRef,
+    HTMLAttributeCollectionEnum_Release,
+    HTMLAttributeCollectionEnum_Next,
+    HTMLAttributeCollectionEnum_Skip,
+    HTMLAttributeCollectionEnum_Reset,
+    HTMLAttributeCollectionEnum_Clone
+};
 
 /* interface IHTMLAttributeCollection */
 static inline HTMLAttributeCollection *impl_from_IHTMLAttributeCollection(IHTMLAttributeCollection *iface)
@@ -7781,10 +7957,24 @@
 static HRESULT WINAPI HTMLAttributeCollection__newEnum(IHTMLAttributeCollection *iface, IUnknown **p)
 {
     HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection(iface);
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    return create_enum_variant(&This->dispex, p);
+    HTMLAttributeCollectionEnum *ret;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    ret = malloc(sizeof(*ret));
+    if(!ret)
+        return E_OUTOFMEMORY;
+
+    ret->IEnumVARIANT_iface.lpVtbl = &HTMLAttributeCollectionEnumVtbl;
+    ret->ref = 1;
+    ret->iter = 0;
+    ret->iter_dispid = DISPID_STARTENUM;
+
+    HTMLAttributeCollection_AddRef(&This->IHTMLAttributeCollection_iface);
+    ret->col = This;
+
+    *p = (IUnknown*)&ret->IEnumVARIANT_iface;
+    return S_OK;
 }
 
 static HRESULT WINAPI HTMLAttributeCollection_item(IHTMLAttributeCollection *iface, VARIANT *name, IDispatch **ppItem)
@@ -7971,141 +8161,6 @@
     HTMLAttributeCollection3_get_length
 };
 
-static inline HTMLAttributeCollection *impl_from_IHTMLAttributeCollection4(IHTMLAttributeCollection4 *iface)
-{
-    return CONTAINING_RECORD(iface, HTMLAttributeCollection, IHTMLAttributeCollection4_iface);
-}
-
-DISPEX_IDISPATCH_IMPL(HTMLAttributeCollection4, IHTMLAttributeCollection4,
-                      impl_from_IHTMLAttributeCollection4(iface)->dispex)
-
-static HRESULT WINAPI HTMLAttributeCollection4_getNamedItemNS(IHTMLAttributeCollection4 *iface, VARIANT *ns,
-                                                              BSTR name, IHTMLDOMAttribute2 **p)
-{
-    HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection4(iface);
-    FIXME("(%p)->(%s %p)\n", This, debugstr_w(name), p);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI HTMLAttributeCollection4_setNamedItemNS(IHTMLAttributeCollection4 *iface,
-        IHTMLDOMAttribute2 *node, IHTMLDOMAttribute2 **p)
-{
-    HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection4(iface);
-    FIXME("(%p)->(%p %p)\n", This, node, p);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI HTMLAttributeCollection4_removeNamedItemNS(IHTMLAttributeCollection4 *iface, VARIANT *ns,
-        BSTR name, IHTMLDOMAttribute2 **p)
-{
-    HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection4(iface);
-    FIXME("(%p)->(%s %p)\n", This, debugstr_w(name), p);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI HTMLAttributeCollection4_getNamedItem(IHTMLAttributeCollection4 *iface, BSTR name,
-        IHTMLDOMAttribute2 **p)
-{
-    HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection4(iface);
-    HTMLDOMAttribute *attr = NULL;
-    nsIDOMAttr *nsattr;
-    nsresult nsres;
-    HRESULT hres;
-
-    TRACE("(%p)->(%s %p)\n", This, debugstr_w(name), p);
-
-    if(This->dom_attrs) {
-        nsAString nsstr;
-        nsAString_Init(&nsstr, name);
-        nsres = nsIDOMMozNamedAttrMap_GetNamedItem(This->dom_attrs, &nsstr, &nsattr);
-        nsAString_Finish(&nsstr);
-        if(NS_FAILED(nsres))
-            return map_nsresult(nsres);
-        if(nsattr) {
-            hres = get_attr_node(nsattr, &attr);
-            nsIDOMAttr_Release(nsattr);
-            if(FAILED(hres))
-               return hres;
-        }
-    }
-
-    *p = attr ? &attr->IHTMLDOMAttribute2_iface : NULL;
-    return S_OK;
-}
-
-static HRESULT WINAPI HTMLAttributeCollection4_setNamedItem(IHTMLAttributeCollection4 *iface,
-        IHTMLDOMAttribute2 *node, IHTMLDOMAttribute2 **p)
-{
-    HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection4(iface);
-    FIXME("(%p)->(%p %p)\n", This, node, p);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI HTMLAttributeCollection4_removeNamedItem(IHTMLAttributeCollection4 *iface,
-        BSTR name, IHTMLDOMAttribute2 **p)
-{
-    HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection4(iface);
-    FIXME("(%p)->(%s %p)\n", This, debugstr_w(name), p);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI HTMLAttributeCollection4_item(IHTMLAttributeCollection4 *iface, LONG index, IHTMLDOMAttribute2 **p)
-{
-    HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection4(iface);
-    HTMLDOMAttribute *attr = NULL;
-    nsIDOMAttr *nsattr;
-    nsresult nsres;
-    HRESULT hres;
-
-    TRACE("(%p)->(%ld %p)\n", This, index, p);
-
-    if(This->dom_attrs) {
-        nsres = nsIDOMMozNamedAttrMap_Item(This->dom_attrs, index, &nsattr);
-        if(NS_FAILED(nsres))
-            return map_nsresult(nsres);
-        if(nsattr) {
-            hres = get_attr_node(nsattr, &attr);
-            nsIDOMAttr_Release(nsattr);
-            if(FAILED(hres))
-               return hres;
-        }
-    }
-
-    *p = attr ? &attr->IHTMLDOMAttribute2_iface : NULL;
-    return S_OK;
-}
-
-static HRESULT WINAPI HTMLAttributeCollection4_get_length(IHTMLAttributeCollection4 *iface, LONG *p)
-{
-    HTMLAttributeCollection *This = impl_from_IHTMLAttributeCollection4(iface);
-    UINT32 length = 0;
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    if(This->dom_attrs)
-        nsIDOMMozNamedAttrMap_GetLength(This->dom_attrs, &length);
-    *p = length;
-    return S_OK;
-}
-
-static const IHTMLAttributeCollection4Vtbl HTMLAttributeCollection4Vtbl = {
-    HTMLAttributeCollection4_QueryInterface,
-    HTMLAttributeCollection4_AddRef,
-    HTMLAttributeCollection4_Release,
-    HTMLAttributeCollection4_GetTypeInfoCount,
-    HTMLAttributeCollection4_GetTypeInfo,
-    HTMLAttributeCollection4_GetIDsOfNames,
-    HTMLAttributeCollection4_Invoke,
-    HTMLAttributeCollection4_getNamedItemNS,
-    HTMLAttributeCollection4_setNamedItemNS,
-    HTMLAttributeCollection4_removeNamedItemNS,
-    HTMLAttributeCollection4_getNamedItem,
-    HTMLAttributeCollection4_setNamedItem,
-    HTMLAttributeCollection4_removeNamedItem,
-    HTMLAttributeCollection4_item,
-    HTMLAttributeCollection4_get_length
-};
-
 static inline HTMLAttributeCollection *HTMLAttributeCollection_from_DispatchEx(DispatchEx *iface)
 {
     return CONTAINING_RECORD(iface, HTMLAttributeCollection, dispex);
@@ -8121,8 +8176,6 @@
         return &This->IHTMLAttributeCollection2_iface;
     if(IsEqualGUID(&IID_IHTMLAttributeCollection3, riid))
         return &This->IHTMLAttributeCollection3_iface;
-    if(IsEqualGUID(&IID_IHTMLAttributeCollection4, riid))
-        return &This->IHTMLAttributeCollection4_iface;
 
     return NULL;
 }
@@ -8132,8 +8185,6 @@
     HTMLAttributeCollection *This = HTMLAttributeCollection_from_DispatchEx(dispex);
     HTMLDOMAttribute *attr;
 
-    if(This->dom_attrs)
-        note_cc_edge((nsISupports*)This->dom_attrs, "dom_attrs", cb);
     LIST_FOR_EACH_ENTRY(attr, &This->attrs, HTMLDOMAttribute, entry)
         note_cc_edge((nsISupports*)&attr->IHTMLDOMAttribute_iface, "attr", cb);
     if(This->elem)
@@ -8143,7 +8194,6 @@
 static void HTMLAttributeCollection_unlink(DispatchEx *dispex)
 {
     HTMLAttributeCollection *This = HTMLAttributeCollection_from_DispatchEx(dispex);
-    unlink_ref(&This->dom_attrs);
     while(!list_empty(&This->attrs)) {
         HTMLDOMAttribute *attr = LIST_ENTRY(list_head(&This->attrs), HTMLDOMAttribute, entry);
 
@@ -8172,46 +8222,6 @@
 
     TRACE("(%p)->(%s %lx %p)\n", This, debugstr_w(name), flags, dispid);
 
-    if(dispex_compat_mode(&This->dispex) >= COMPAT_MODE_IE9) {
-        nsIDOMAttr *nsattr, *iter;
-        UINT32 length, i;
-        nsAString nsstr;
-        nsresult nsres;
-
-        if(!This->dom_attrs)
-            return DISP_E_UNKNOWNNAME;
-        nsIDOMMozNamedAttrMap_GetLength(This->dom_attrs, &length);
-
-        if(name[0] >= '0' && name[0] <= '9') {
-            WCHAR *end_ptr;
-            i = wcstoul(name, &end_ptr, 10);
-            if(!*end_ptr) {
-                if(i < length) {
-                    *dispid = MSHTML_DISPID_CUSTOM_MIN + i;
-                    return S_OK;
-                }
-            }
-        }
-
-        nsAString_Init(&nsstr, name);
-        nsres = nsIDOMMozNamedAttrMap_GetNamedItem(This->dom_attrs, &nsstr, &nsattr);
-        nsAString_Finish(&nsstr);
-        if(NS_FAILED(nsres) || !nsattr)
-            return DISP_E_UNKNOWNNAME;
-
-        for(i = 0; i < length; i++) {
-            nsres = nsIDOMMozNamedAttrMap_Item(This->dom_attrs, i, &iter);
-            assert(nsres == NS_OK);
-            nsIDOMAttr_Release(iter);
-            if(iter == nsattr)
-                break;
-        }
-
-        assert(i < length);
-        *dispid = MSHTML_DISPID_CUSTOM_MIN + i;
-        return S_OK;
-    }
-
     hres = get_attr_dispid_by_name(This, name, dispid);
     if(FAILED(hres))
         return hres;
@@ -8221,8 +8231,6 @@
         return hres;
     IHTMLDOMAttribute_Release(&attr->IHTMLDOMAttribute_iface);
 
-    /* Even though this breaks DISPID rules where the same name must return the same DISPID, because the pos can change
-     * as attributes are removed (and re-added), it's how native works (see test_attr_collection_disp in tests). */
     *dispid = MSHTML_DISPID_CUSTOM_MIN+pos;
     return S_OK;
 }
@@ -8240,20 +8248,6 @@
         DWORD pos;
 
         pos = id-MSHTML_DISPID_CUSTOM_MIN;
-
-        if(dispex_compat_mode(&This->dispex) >= COMPAT_MODE_IE9) {
-            IHTMLDOMAttribute2 *attr;
-            HRESULT hres;
-
-            hres = IHTMLAttributeCollection4_item(&This->IHTMLAttributeCollection4_iface, pos, &attr);
-            if(FAILED(hres))
-                return hres;
-            if(!attr)
-                return DISP_E_MEMBERNOTFOUND;
-            V_VT(res) = VT_DISPATCH;
-            V_DISPATCH(res) = (IDispatch *)attr;
-            return S_OK;
-        }
 
         LIST_FOR_EACH_ENTRY(iter, &This->attrs, HTMLDOMAttribute, entry) {
             if(!pos) {
@@ -8275,26 +8269,6 @@
     }
 }
 
-static ULONG HTMLAttributeCollection_collection_len(DispatchEx *dispex)
-{
-    HTMLAttributeCollection *This = HTMLAttributeCollection_from_DispatchEx(dispex);
-    LONG idx = -1;
-
-    return SUCCEEDED(get_attr_dispid_by_idx(This, &idx, NULL)) ? idx : 0;
-}
-
-static HRESULT HTMLAttributeCollection_collection_item(DispatchEx *dispex, ULONG index, IDispatch **p)
-{
-    HTMLAttributeCollection *This = HTMLAttributeCollection_from_DispatchEx(dispex);
-    IHTMLDOMAttribute *attr;
-    HRESULT hres;
-
-    hres = IHTMLAttributeCollection3_item(&This->IHTMLAttributeCollection3_iface, index, &attr);
-    if(SUCCEEDED(hres))
-        *p = (IDispatch*)attr;
-    return hres;
-}
-
 static const dispex_static_data_vtbl_t HTMLAttributeCollection_dispex_vtbl = {
     .query_interface  = HTMLAttributeCollection_query_interface,
     .destructor       = HTMLAttributeCollection_destructor,
@@ -8303,27 +8277,20 @@
     .get_dispid       = HTMLAttributeCollection_get_dispid,
     .get_prop_desc    = dispex_index_prop_desc,
     .invoke           = HTMLAttributeCollection_invoke,
-    .collection_len   = HTMLAttributeCollection_collection_len,
-    .collection_item  = HTMLAttributeCollection_collection_item,
 };
 
-static void NamedNodeMap_init_dispex_info(dispex_data_t *info, compat_mode_t mode)
-{
-    if(mode >= COMPAT_MODE_IE9) {
-        dispex_info_add_interface(info, IHTMLAttributeCollection4_tid, NULL);
-    }else if(mode == COMPAT_MODE_IE8) {
-        dispex_info_add_interface(info, IHTMLAttributeCollection3_tid, NULL);
-    }else {
-        dispex_info_add_interface(info, IHTMLAttributeCollection_tid, NULL);
-        dispex_info_add_interface(info, IHTMLAttributeCollection2_tid, NULL);
-    }
-}
+const tid_t NamedNodeMap_iface_tids[] = {
+    IHTMLAttributeCollection_tid,
+    IHTMLAttributeCollection2_tid,
+    IHTMLAttributeCollection3_tid,
+    0
+};
 
 dispex_static_data_t NamedNodeMap_dispex = {
-    .id         = OBJID_NamedNodeMap,
+    .id         = PROT_NamedNodeMap,
     .vtbl       = &HTMLAttributeCollection_dispex_vtbl,
     .disp_tid   = DispHTMLAttributeCollection_tid,
-    .init_info  = NamedNodeMap_init_dispex_info,
+    .iface_tids = NamedNodeMap_iface_tids,
     .js_flags   = HOSTOBJ_VOLATILE_PROPS
 };
 
@@ -8344,7 +8311,6 @@
     This->attrs->IHTMLAttributeCollection_iface.lpVtbl = &HTMLAttributeCollectionVtbl;
     This->attrs->IHTMLAttributeCollection2_iface.lpVtbl = &HTMLAttributeCollection2Vtbl;
     This->attrs->IHTMLAttributeCollection3_iface.lpVtbl = &HTMLAttributeCollection3Vtbl;
-    This->attrs->IHTMLAttributeCollection4_iface.lpVtbl = &HTMLAttributeCollection4Vtbl;
 
     IHTMLDOMNode_AddRef(&This->node.IHTMLDOMNode_iface);
     This->attrs->elem = This;
@@ -8352,9 +8318,6 @@
     init_dispatch(&This->attrs->dispex, &NamedNodeMap_dispex, This->node.doc->script_global,
                   dispex_compat_mode(&This->node.event_target.dispex));
 
-    if(This->dom_element)
-        nsIDOMElement_GetAttributes(This->dom_element, &This->attrs->dom_attrs);
-
     *ac = This->attrs;
     IHTMLAttributeCollection_AddRef(&This->attrs->IHTMLAttributeCollection_iface);
     return S_OK;
