diff --git a/dlls/ddraw/tests/ddraw7.c b/dlls/ddraw/tests/ddraw7.c
--- a/dlls/ddraw/tests/ddraw7.c
+++ b/dlls/ddraw/tests/ddraw7.c
@@ -21,7 +21,6 @@
 #define COBJMACROS
 
 #include <stdbool.h>
-#include <stdint.h>
 #include "wine/test.h"
 #include <limits.h>
 #include <math.h>
@@ -31,8 +30,7 @@
 
 static HRESULT (WINAPI *pDirectDrawCreateEx)(GUID *guid, void **ddraw, REFIID iid, IUnknown *outer_unknown);
 static BOOL is_ddraw64 = sizeof(DWORD) != sizeof(DWORD *);
-static DEVMODEA registry_mode;
-static HWND foreground;
+static DEVMODEW registry_mode;
 static const GUID *hw_device_guid = &IID_IDirect3DHALDevice;
 
 static HRESULT (WINAPI *pDwmIsCompositionEnabled)(BOOL *);
@@ -112,7 +110,7 @@
             && compare_uint((c1 >> 24) & 0xff, (c2 >> 24) & 0xff, max_diff);
 }
 
-static BOOL compare_mode_rect(const DEVMODEA *mode1, const DEVMODEA *mode2)
+static BOOL compare_mode_rect(const DEVMODEW *mode1, const DEVMODEW *mode2)
 {
     return mode1->dmPosition.x == mode2->dmPosition.x
             && mode1->dmPosition.y == mode2->dmPosition.y
@@ -160,64 +158,6 @@
             DispatchMessageA(&msg);
         diff = time - GetTickCount();
     }
-}
-
-static HWND create_foreground_window(void)
-{
-    for (UINT retries = 5; retries; retries--)
-    {
-        HWND hwnd;
-        BOOL ret;
-
-        hwnd = CreateWindowW(L"static", NULL, WS_POPUP | WS_VISIBLE, 0, 0, 5, 5, NULL, NULL, NULL, NULL);
-        ok(hwnd != NULL, "CreateWindowW failed, error %lu\n", GetLastError());
-        flush_events();
-
-        if (GetForegroundWindow() == hwnd)
-            return hwnd;
-        ret = DestroyWindow(hwnd);
-        ok(ret, "DestroyWindow failed, error %lu\n", GetLastError());
-        flush_events();
-    }
-
-    ok(0, "Failed to create foreground window\n");
-    return NULL;
-}
-
-static LRESULT CALLBACK foreground_window_wndproc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam)
-{
-    if (msg == WM_USER)
-        SetForegroundWindow((HWND)wparam);
-    if (msg == WM_CLOSE)
-        PostQuitMessage(0);
-    return DefWindowProcW(hwnd, msg, wparam, lparam);
-}
-
-static DWORD WINAPI foreground_window_thread(void *arg)
-{
-    HANDLE event = arg;
-    MSG msg;
-
-    foreground = create_foreground_window();
-    SetWindowLongPtrW(foreground, GWLP_WNDPROC, (LONG_PTR)foreground_window_wndproc);
-    SetEvent(event);
-
-    while (GetMessageW(&msg, NULL, 0, 0))
-    {
-        if (msg.message == WM_QUIT) break;
-        TranslateMessage(&msg);
-        DispatchMessageW(&msg);
-    }
-
-    return 0;
-}
-
-static void start_foreground_window_thread(void)
-{
-    HANDLE event = CreateEventA(NULL, FALSE, FALSE, NULL);
-    CloseHandle(CreateThread(NULL, 0, foreground_window_thread, event, 0, NULL));
-    WaitForSingleObject(event, 10000);
-    CloseHandle(event);
 }
 
 static BOOL ddraw_get_identifier(IDirectDraw7 *ddraw, DDDEVICEIDENTIFIER2 *identifier)
@@ -373,20 +313,20 @@
 }
 
 /* Free original_modes after finished using it */
-static BOOL save_display_modes(DEVMODEA **original_modes, unsigned int *display_count)
+static BOOL save_display_modes(DEVMODEW **original_modes, unsigned int *display_count)
 {
     unsigned int number, size = 2, count = 0, index = 0;
-    DISPLAY_DEVICEA display_device;
-    DEVMODEA *modes, *tmp;
+    DISPLAY_DEVICEW display_device;
+    DEVMODEW *modes, *tmp;
 
     if (!(modes = malloc(size * sizeof(*modes))))
         return FALSE;
 
     display_device.cb = sizeof(display_device);
-    while (EnumDisplayDevicesA(NULL, index++, &display_device, 0))
+    while (EnumDisplayDevicesW(NULL, index++, &display_device, 0))
     {
         /* Skip software devices */
-        if (sscanf(display_device.DeviceName, "\\\\.\\DISPLAY%u", &number) != 1)
+        if (swscanf(display_device.DeviceName, L"\\\\.\\DISPLAY%u", &number) != 1)
             continue;
 
         if (!(display_device.StateFlags & DISPLAY_DEVICE_ATTACHED_TO_DESKTOP))
@@ -405,13 +345,13 @@
 
         memset(&modes[count], 0, sizeof(modes[count]));
         modes[count].dmSize = sizeof(modes[count]);
-        if (!EnumDisplaySettingsA(display_device.DeviceName, ENUM_CURRENT_SETTINGS, &modes[count]))
+        if (!EnumDisplaySettingsW(display_device.DeviceName, ENUM_CURRENT_SETTINGS, &modes[count]))
         {
             free(modes);
             return FALSE;
         }
 
-        strcpy((char *)modes[count++].dmDeviceName, display_device.DeviceName);
+        lstrcpyW(modes[count++].dmDeviceName, display_device.DeviceName);
     }
 
     *original_modes = modes;
@@ -419,19 +359,19 @@
     return TRUE;
 }
 
-static BOOL restore_display_modes(DEVMODEA *modes, unsigned int count)
+static BOOL restore_display_modes(DEVMODEW *modes, unsigned int count)
 {
     unsigned int index;
     LONG ret;
 
     for (index = 0; index < count; ++index)
     {
-        ret = ChangeDisplaySettingsExA((char *)modes[index].dmDeviceName, &modes[index], NULL,
+        ret = ChangeDisplaySettingsExW(modes[index].dmDeviceName, &modes[index], NULL,
                 CDS_UPDATEREGISTRY | CDS_NORESET, NULL);
         if (ret != DISP_CHANGE_SUCCESSFUL)
             return FALSE;
     }
-    ret = ChangeDisplaySettingsExA(NULL, NULL, NULL, 0, NULL);
+    ret = ChangeDisplaySettingsExW(NULL, NULL, NULL, 0, NULL);
     return ret == DISP_CHANGE_SUCCESSFUL;
 }
 
@@ -1237,8 +1177,6 @@
     window = CreateWindowA("static", "ddraw_test", WS_OVERLAPPEDWINDOW,
             10, 10, 640, 480, 0, 0, 0, 0);
     ShowWindow(window, SW_SHOW);
-    flush_events();
-
     ddraw = create_ddraw();
     ok(!!ddraw, "Failed to create a ddraw object.\n");
 
@@ -2776,7 +2714,7 @@
     GetClientRect(window, &r);
     todo_wine ok(!EqualRect(&r, &fullscreen_rect), "Client rect and window rect are equal.\n");
 
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
 
     tmp = GetWindowLongA(window, GWL_STYLE);
@@ -2864,7 +2802,7 @@
     ok(tmp == exstyle, "Expected window extended style %#lx, got %#lx.\n", exstyle, tmp);
 
     ShowWindow(window, SW_SHOW);
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     SetActiveWindow(window);
     ok(GetActiveWindow() == window, "Unexpected active window.\n");
@@ -2936,7 +2874,7 @@
     expected_style = exstyle | WS_EX_TOPMOST;
     todo_wine ok(tmp == expected_style, "Expected window extended style %#lx, got %#lx.\n", expected_style, tmp);
 
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     tmp = GetWindowLongA(window, GWL_STYLE);
     expected_style = style | WS_VISIBLE | WS_MINIMIZE;
@@ -3094,7 +3032,7 @@
 
 static void test_coop_level_mode_set(void)
 {
-    DEVMODEA *original_modes = NULL, devmode, devmode2;
+    DEVMODEW *original_modes = NULL, devmode, devmode2;
     unsigned int display_count = 0;
     IDirectDrawSurface7 *primary;
     RECT registry_rect, ddraw_rect, user32_rect, r;
@@ -3189,11 +3127,11 @@
 
     memset(&devmode, 0, sizeof(devmode));
     devmode.dmSize = sizeof(devmode);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode, &registry_mode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode, &registry_mode), "Got a different mode.\n");
 
     ret = save_display_modes(&original_modes, &display_count);
@@ -3224,7 +3162,7 @@
     devmode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
     devmode.dmPelsWidth = param.user32_width;
     devmode.dmPelsHeight = param.user32_height;
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
 
     ddraw = create_ddraw();
@@ -3329,7 +3267,7 @@
 
     hr = IDirectDrawSurface7_IsLost(primary);
     ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
     hr = IDirectDrawSurface7_IsLost(primary);
     ok(hr == DDERR_SURFACELOST, "Got unexpected hr %#lx.\n", hr);
@@ -3346,14 +3284,14 @@
 
     expect_messages = exclusive_focus_loss_messages;
     focus_test_ddraw = ddraw;
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     ok(!expect_messages->message, "Expected message %#x, but didn't receive it.\n", expect_messages->message);
     focus_test_ddraw = NULL;
 
     memset(&devmode, 0, sizeof(devmode));
     devmode.dmSize = sizeof(devmode);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
     ok(ret, "Failed to get display mode.\n");
     ok(devmode.dmPelsWidth == registry_mode.dmPelsWidth
             && devmode.dmPelsHeight == registry_mode.dmPelsHeight, "Got unexpected screen size %lux%lu.\n",
@@ -3366,7 +3304,7 @@
     GetWindowRect(window, &r);
     ok(EqualRect(&r, &ddraw_rect), "Expected %s, got %s.\n", wine_dbgstr_rect(&ddraw_rect),
             wine_dbgstr_rect(&r));
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
     ok(ret, "Failed to get display mode.\n");
     ok(devmode.dmPelsWidth == param.ddraw_width
             && devmode.dmPelsHeight == param.ddraw_height, "Got unexpected screen size %lux%lu.\n",
@@ -3439,7 +3377,7 @@
     expect_messages = NULL;
 
     /* For Wine. */
-    change_ret = ChangeDisplaySettingsA(NULL, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(NULL, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
     flush_events();
 
@@ -3520,7 +3458,7 @@
     devmode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
     devmode.dmPelsWidth = param.user32_width;
     devmode.dmPelsHeight = param.user32_height;
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
     hr = IDirectDrawSurface7_IsLost(primary);
     todo_wine ok(hr == DDERR_SURFACELOST, "Got unexpected hr %#lx.\n", hr);
@@ -3626,14 +3564,14 @@
     ok(expect_messages->message == WM_PAINT, "Unexpected WM_PAINT.\n");
     expect_messages = NULL;
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
     ok(ret, "Failed to get display mode.\n");
     ok(devmode.dmPelsWidth == registry_mode.dmPelsWidth
             && devmode.dmPelsHeight == registry_mode.dmPelsHeight,
             "Expected resolution %lux%lu, got %lux%lu.\n",
             registry_mode.dmPelsWidth, registry_mode.dmPelsHeight,
             devmode.dmPelsWidth, devmode.dmPelsHeight);
-    change_ret = ChangeDisplaySettingsA(NULL, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(NULL, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
 
     memset(&ddsd, 0, sizeof(ddsd));
@@ -3710,7 +3648,7 @@
     devmode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
     devmode.dmPelsWidth = param.user32_width;
     devmode.dmPelsHeight = param.user32_height;
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
     hr = IDirectDrawSurface7_IsLost(primary);
     todo_wine ok(hr == DDERR_SURFACELOST, "Got unexpected hr %#lx.\n", hr);
@@ -3816,14 +3754,14 @@
     ok(expect_messages->message == WM_PAINT, "Unexpected WM_PAINT.\n");
     expect_messages = NULL;
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
     ok(ret, "Failed to get display mode.\n");
     ok(devmode.dmPelsWidth == registry_mode.dmPelsWidth
             && devmode.dmPelsHeight == registry_mode.dmPelsHeight,
             "Expected resolution %lux%lu, got %lux%lu.\n",
             registry_mode.dmPelsWidth, registry_mode.dmPelsHeight,
             devmode.dmPelsWidth, devmode.dmPelsHeight);
-    change_ret = ChangeDisplaySettingsA(NULL, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(NULL, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
 
     memset(&ddsd, 0, sizeof(ddsd));
@@ -4005,8 +3943,8 @@
     devmode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
     devmode.dmPelsWidth = param.user32_width;
     devmode.dmPelsHeight = param.user32_height;
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsW failed with %ld.\n", change_ret);
 
     ddraw = create_ddraw();
     ok(!!ddraw, "Failed to create a ddraw object.\n");
@@ -4015,15 +3953,15 @@
 
     memset(&devmode2, 0, sizeof(devmode2));
     devmode2.dmSize = sizeof(devmode2);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &registry_mode), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
 
     /* Test that no mode restorations if no mode changes happened with fullscreen ddraw objects */
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsW failed with %ld.\n", change_ret);
 
     ddraw = create_ddraw();
     ok(!!ddraw, "Failed to create a ddraw object.\n");
@@ -4034,8 +3972,8 @@
     ref = IDirectDraw7_Release(ddraw);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &registry_mode), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
@@ -4047,17 +3985,17 @@
     hr = set_display_mode(ddraw, registry_mode.dmPelsWidth, registry_mode.dmPelsHeight);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsW failed with %ld.\n", change_ret);
 
     ref = IDirectDraw7_Release(ddraw);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &devmode), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
@@ -4068,17 +4006,17 @@
     hr = set_display_mode(ddraw, param.ddraw_width, param.ddraw_height);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsW failed with %ld.\n", change_ret);
 
     hr = IDirectDraw7_RestoreDisplayMode(ddraw);
     ok(hr == DD_OK, "RestoreDisplayMode failed, hr %#lx.\n", hr);
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &devmode), "Got a different mode.\n");
 
     ref = IDirectDraw7_Release(ddraw);
@@ -4096,9 +4034,9 @@
 
 static void test_coop_level_mode_set_multi(void)
 {
-    DEVMODEA old_devmode, devmode, devmode2, devmode3, *original_modes = NULL;
+    DEVMODEW old_devmode, devmode, devmode2, devmode3, *original_modes = NULL;
     unsigned int mode_idx = 0, display_idx, display_count = 0;
-    char second_monitor_name[CCHDEVICENAME];
+    WCHAR second_monitor_name[CCHDEVICENAME];
     IDirectDraw7 *ddraw1, *ddraw2;
     LONG change_ret;
     UINT w, h;
@@ -4109,11 +4047,11 @@
 
     memset(&devmode, 0, sizeof(devmode));
     devmode.dmSize = sizeof(devmode);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode, &registry_mode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode, &registry_mode), "Got a different mode.\n");
 
     ret = save_display_modes(&original_modes, &display_count);
@@ -4306,18 +4244,18 @@
     {
         if (original_modes[display_idx].dmPosition.x || original_modes[display_idx].dmPosition.y)
         {
-            strcpy(second_monitor_name, (char *)original_modes[display_idx].dmDeviceName);
+            lstrcpyW(second_monitor_name, original_modes[display_idx].dmDeviceName);
             break;
         }
     }
-    ok(strlen(second_monitor_name), "Got an empty second monitor name.\n");
+    ok(lstrlenW(second_monitor_name), "Got an empty second monitor name.\n");
     memset(&old_devmode, 0, sizeof(old_devmode));
     old_devmode.dmSize = sizeof(old_devmode);
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &old_devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &old_devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
 
     devmode = old_devmode;
-    while (EnumDisplaySettingsA(second_monitor_name, mode_idx++, &devmode))
+    while (EnumDisplaySettingsW(second_monitor_name, mode_idx++, &devmode))
     {
         if (devmode.dmPelsWidth != old_devmode.dmPelsWidth
                 || devmode.dmPelsHeight != old_devmode.dmPelsHeight)
@@ -4333,13 +4271,13 @@
     hr = IDirectDraw7_SetCooperativeLevel(ddraw1, window, DDSCL_EXCLUSIVE | DDSCL_FULLSCREEN);
     ok(hr == DD_OK, "SetCooperativeLevel failed, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     memset(&devmode2, 0, sizeof(devmode2));
     devmode2.dmSize = sizeof(devmode2);
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     if (compare_mode_rect(&devmode2, &old_devmode))
     {
         skip("Failed to change display settings of the second monitor.\n");
@@ -4355,8 +4293,8 @@
 
     memset(&devmode3, 0, sizeof(devmode3));
     devmode3.dmSize = sizeof(devmode3);
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode3);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode3);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode3, &devmode2), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
@@ -4368,17 +4306,17 @@
     hr = set_display_mode(ddraw1, 800, 600);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     ref = IDirectDraw7_Release(ddraw1);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
@@ -4389,17 +4327,17 @@
     hr = set_display_mode(ddraw1, 800, 600);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     hr = IDirectDraw7_RestoreDisplayMode(ddraw1);
     ok(hr == DD_OK, "RestoreDisplayMode failed, hr %#lx.\n", hr);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
 
     ref = IDirectDraw7_Release(ddraw1);
@@ -4413,20 +4351,20 @@
     hr = set_display_mode(ddraw1, 800, 600);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL,
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL,
             CDS_UPDATEREGISTRY | CDS_NORESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     ref = IDirectDraw7_Release(ddraw1);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(devmode2.dmPelsWidth == devmode.dmPelsWidth && devmode2.dmPelsHeight == devmode.dmPelsHeight,
             "Expected resolution %lux%lu, got %lux%lu.\n", devmode.dmPelsWidth, devmode.dmPelsHeight,
             devmode2.dmPelsWidth, devmode2.dmPelsHeight);
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(devmode2.dmPelsWidth == devmode.dmPelsWidth && devmode2.dmPelsHeight == devmode.dmPelsHeight,
             "Expected resolution %lux%lu, got %lux%lu.\n", devmode.dmPelsWidth, devmode.dmPelsHeight,
             devmode2.dmPelsWidth, devmode2.dmPelsHeight);
@@ -4444,17 +4382,17 @@
     hr = set_display_mode(ddraw2, 640, 480);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     hr = IDirectDraw7_RestoreDisplayMode(ddraw2);
     ok(hr == DD_OK, "RestoreDisplayMode failed, hr %#lx.\n", hr);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
 
     ref = IDirectDraw7_Release(ddraw2);
@@ -4475,17 +4413,17 @@
     hr = set_display_mode(ddraw2, 640, 480);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     ref = IDirectDraw7_Release(ddraw2);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
 
     ref = IDirectDraw7_Release(ddraw1);
@@ -5755,7 +5693,6 @@
         ok(SUCCEEDED(hr), "Failed to end scene, hr %#lx.\n", hr);
 
         color = get_surface_color(rt, 310, 240);
-        todo_wine_if(i == 1)
         ok(compare_color(color, tests[i].color_left, 1),
                 "Expected left color 0x%08x, got 0x%08x, case %u.\n", tests[i].color_left, color, i);
         color = get_surface_color(rt, 330, 240);
@@ -6021,7 +5958,7 @@
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
 
     /* Exclusive with window not active. */
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     activateapp_testdata.received = FALSE;
     hr = IDirectDraw7_SetCooperativeLevel(ddraw, window, DDSCL_EXCLUSIVE | DDSCL_FULLSCREEN);
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
@@ -6030,7 +5967,7 @@
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
 
     /* Normal with window not active, then exclusive with the same window. */
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     activateapp_testdata.received = FALSE;
     hr = IDirectDraw7_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
@@ -6042,7 +5979,7 @@
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
 
     /* Recursive set of DDSCL_EXCLUSIVE | DDSCL_FULLSCREEN. */
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     activateapp_testdata.received = FALSE;
     activateapp_testdata.ddraw = ddraw;
     activateapp_testdata.window = window;
@@ -6064,7 +6001,7 @@
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
 
     /* Setting DDSCL_NORMAL with recursive invocation. */
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     activateapp_testdata.received = FALSE;
     activateapp_testdata.ddraw = ddraw;
     activateapp_testdata.window = window;
@@ -10559,7 +10496,7 @@
         ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
     }
 
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     hr = IDirectDraw7_TestCooperativeLevel(ddraw);
     ok(hr == DDERR_NOEXCLUSIVEMODE, "Got unexpected hr %#lx.\n", hr);
@@ -10661,7 +10598,7 @@
         ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
     }
 
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     hr = IDirectDraw7_TestCooperativeLevel(ddraw);
     ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
@@ -17546,7 +17483,7 @@
     hr = IDirectDraw7_CreateSurface(killfocus_ddraw, &surface_desc, &killfocus_surface, NULL);
     ok(SUCCEEDED(hr), "Failed to create surface, hr %#lx.\n", hr);
 
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     ok(!killfocus_ddraw, "WM_KILLFOCUS was not received.\n");
 
     DestroyWindow(window);
@@ -18219,59 +18156,46 @@
 
 static void test_caps(void)
 {
+    DWORD caps_never, caps_always, caps_hal;
     DDCAPS hal_caps, hel_caps;
     IDirectDraw7 *ddraw;
-    HWND window;
     HRESULT hr;
     BOOL no3d;
 
-    static const struct
-    {
-        unsigned int depth;
-        DWORD flag;
-        DWORD mask;
-    }
-    depth_caps[] =
-    {
-        {16, DDBD_16, 0xffff},
-        {24, DDBD_24, 0xffffff},
-        {32, DDBD_32, 0xffffffff},
-    };
-
-    static const DWORD caps_hel = DDSCAPS_FLIP
-            | DDSCAPS_OFFSCREENPLAIN
-            | DDSCAPS_PALETTE
-            | DDSCAPS_PRIMARYSURFACE
-            | DDSCAPS_TEXTURE
-            | DDSCAPS_ZBUFFER
-            | DDSCAPS_MIPMAP;
-
-    static const DWORD caps_never = DDSCAPS_RESERVED1
+    caps_never = DDSCAPS_RESERVED1
+            | DDSCAPS_ALPHA
             | DDSCAPS_PRIMARYSURFACELEFT
+            | DDSCAPS_SYSTEMMEMORY
+            | DDSCAPS_VISIBLE
             | DDSCAPS_WRITEONLY
             | DDSCAPS_LIVEVIDEO
+            | DDSCAPS_HWCODEC
+            | DDSCAPS_MODEX
             | DDSCAPS_RESERVED2
             | 0x01000000u
             | 0x02000000u
             | DDSCAPS_ALLOCONLOAD
             | DDSCAPS_VIDEOPORT
+            | DDSCAPS_STANDARDVGAMODE
             | DDSCAPS_OPTIMIZED;
 
-    static const DWORD caps_always = DDSCAPS_BACKBUFFER
-            | DDSCAPS_FLIP
+    caps_always = DDSCAPS_FLIP
             | DDSCAPS_OFFSCREENPLAIN
             | DDSCAPS_PRIMARYSURFACE
             | DDSCAPS_TEXTURE
+            | DDSCAPS_ZBUFFER
+            | DDSCAPS_MIPMAP;
+
+    caps_hal = DDSCAPS_BACKBUFFER
+            | DDSCAPS_COMPLEX
+            | DDSCAPS_FRONTBUFFER
             | DDSCAPS_3DDEVICE
-            | DDSCAPS_ZBUFFER
             | DDSCAPS_VIDEOMEMORY
-            | DDSCAPS_MIPMAP;
-
-    window = create_window();
+            | DDSCAPS_LOCALVIDMEM
+            | DDSCAPS_NONLOCALVIDMEM;
+
     ddraw = create_ddraw();
     ok(!!ddraw, "Failed to create a ddraw object.\n");
-    hr = IDirectDraw_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
-    ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
 
     memset(&hal_caps, 0, sizeof(hal_caps));
     memset(&hel_caps, 0, sizeof(hel_caps));
@@ -18290,51 +18214,14 @@
     if (hal_caps.ddsCaps.dwCaps)
     {
         ok(!(hal_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
-        todo_wine_if(no3d) ok(!(~hal_caps.ddsCaps.dwCaps & caps_always),
+        ok(!(~hal_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
+        todo_wine_if(no3d) ok(!(~hal_caps.ddsCaps.dwCaps & caps_hal),
                 "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
     }
-    todo_wine ok(hel_caps.ddsCaps.dwCaps == caps_hel, "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
-
-    ok(!(hal_caps.dwZBufferBitDepths & ~(DDBD_16 | DDBD_24 | DDBD_32)),
-            "Got HAL depth caps %#lx.\n", hal_caps.dwZBufferBitDepths);
-    todo_wine ok(hel_caps.dwZBufferBitDepths == DDBD_16, "Got HEL depth caps %#lx.\n", hel_caps.dwZBufferBitDepths);
-
-    for (unsigned int i = 0; i < ARRAY_SIZE(depth_caps); ++i)
-    {
-        bool supported = (hal_caps.dwZBufferBitDepths & depth_caps[i].flag);
-        IDirectDrawSurface7 *surface;
-        DDSURFACEDESC2 desc =
-        {
-            .dwSize = sizeof(DDSURFACEDESC2),
-            .dwFlags = DDSD_CAPS | DDSD_PIXELFORMAT | DDSD_WIDTH | DDSD_HEIGHT,
-            .ddsCaps.dwCaps = DDSCAPS_ZBUFFER,
-            .ddpfPixelFormat.dwSize = sizeof(DDPIXELFORMAT),
-            .ddpfPixelFormat.dwFlags = DDPF_ZBUFFER,
-            .ddpfPixelFormat.dwZBufferBitDepth = depth_caps[i].depth,
-            .ddpfPixelFormat.dwZBitMask = depth_caps[i].mask,
-            .dwWidth = 64,
-            .dwHeight = 64,
-        };
-
-        winetest_push_context("depth %u", depth_caps[i].depth);
-
-        /* dwZBufferBitDepths sometimes reports false negatives,
-         * but it has not been known to report false positives. */
-        hr = IDirectDraw7_CreateSurface(ddraw, &desc, &surface, NULL);
-        ok(hr == S_OK || (!supported && hr == DDERR_INVALIDPIXELFORMAT), "Got hr %#lx.\n", hr);
-
-        if (hr == S_OK)
-        {
-            hr = IDirectDrawSurface7_GetSurfaceDesc(surface, &desc);
-            ok(hr == S_OK, "Got hr %#lx.\n", hr);
-            ok(desc.ddsCaps.dwCaps == (DDSCAPS_VIDEOMEMORY | DDSCAPS_LOCALVIDMEM | DDSCAPS_ZBUFFER)
-                    || (!supported && desc.ddsCaps.dwCaps == (DDSCAPS_SYSTEMMEMORY | DDSCAPS_ZBUFFER)),
-                    "Got caps %#lx.\n", desc.ddsCaps.dwCaps);
-            IDirectDrawSurface7_Release(surface);
-        }
-
-        winetest_pop_context();
-    }
+    ok(!(hel_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+    ok(!(~hel_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+    todo_wine_if(!no3d) ok(!(hel_caps.ddsCaps.dwCaps & caps_hal),
+            "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
 
     IDirectDraw7_Release(ddraw);
 
@@ -18357,17 +18244,15 @@
                 hel_caps.ddsOldCaps.dwCaps, hel_caps.ddsCaps.dwCaps);
 
         ok(!(hal_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
-        todo_wine_if(no3d) ok(!(~hal_caps.ddsCaps.dwCaps & caps_always),
+        ok(!(~hal_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
+        todo_wine_if(no3d) ok(!(~hal_caps.ddsCaps.dwCaps & caps_hal),
                 "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
-        todo_wine ok(!(hel_caps.ddsCaps.dwCaps & ~(DDSCAPS_STANDARDVGAMODE | DDSCAPS_MODEX)),
-                "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+        todo_wine ok(!hel_caps.ddsCaps.dwCaps, "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
 
         IDirectDraw7_Release(ddraw);
     }
 
     hr = pDirectDrawCreateEx((GUID *)DDCREATE_EMULATIONONLY, (void **)&ddraw, &IID_IDirectDraw7, NULL);
-    ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
-    hr = IDirectDraw7_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
     ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
 
     memset(&hal_caps, 0, sizeof(hal_caps));
@@ -18383,50 +18268,13 @@
             "Got unexpected caps %#lx, expected %#lx.\n",
             hel_caps.ddsOldCaps.dwCaps, hel_caps.ddsCaps.dwCaps);
 
-    todo_wine ok(!(hal_caps.ddsCaps.dwCaps & ~(DDSCAPS_STANDARDVGAMODE | DDSCAPS_MODEX)),
-            "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
-    todo_wine ok(hel_caps.ddsCaps.dwCaps == caps_hel, "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
-
-    for (unsigned int i = 0; i < ARRAY_SIZE(depth_caps); ++i)
-    {
-        IDirectDrawSurface7 *surface;
-        DDSURFACEDESC2 desc =
-        {
-            .dwSize = sizeof(DDSURFACEDESC2),
-            .dwFlags = DDSD_CAPS | DDSD_PIXELFORMAT | DDSD_WIDTH | DDSD_HEIGHT,
-            .ddsCaps.dwCaps = DDSCAPS_ZBUFFER,
-            .ddpfPixelFormat.dwSize = sizeof(DDPIXELFORMAT),
-            .ddpfPixelFormat.dwFlags = DDPF_ZBUFFER,
-            .ddpfPixelFormat.dwZBufferBitDepth = depth_caps[i].depth,
-            .ddpfPixelFormat.dwZBitMask = depth_caps[i].mask,
-            .dwWidth = 64,
-            .dwHeight = 64,
-        };
-
-        winetest_push_context("depth %u", depth_caps[i].depth);
-
-        hr = IDirectDraw7_CreateSurface(ddraw, &desc, &surface, NULL);
-        if (depth_caps[i].depth == 16 || depth_caps[i].depth == 32)
-            ok(hr == S_OK, "Got hr %#lx.\n", hr);
-        else
-            todo_wine ok(hr == DDERR_INVALIDPIXELFORMAT, "Got hr %#lx.\n", hr);
-
-        if (hr == S_OK)
-        {
-            hr = IDirectDrawSurface7_GetSurfaceDesc(surface, &desc);
-            ok(hr == S_OK, "Got hr %#lx.\n", hr);
-            todo_wine_if (depth_caps[i].depth != 32)
-                ok(desc.ddsCaps.dwCaps == (DDSCAPS_SYSTEMMEMORY | DDSCAPS_ZBUFFER),
-                        "Got caps %#lx.\n", desc.ddsCaps.dwCaps);
-            IDirectDrawSurface7_Release(surface);
-        }
-
-        winetest_pop_context();
-    }
+    todo_wine ok(!hal_caps.ddsCaps.dwCaps, "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
+    ok(!(hel_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+    ok(!(~hel_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+    todo_wine_if(!no3d) ok(!(hel_caps.ddsCaps.dwCaps & caps_hal),
+            "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
 
     IDirectDraw7_Release(ddraw);
-
-    DestroyWindow(window);
 }
 
 static void test_d32_support(void)
@@ -18468,6 +18316,8 @@
             "Got unexpected dwZBufferBitDepth %lu.\n", surface_desc.ddpfPixelFormat.dwZBufferBitDepth);
     ok(surface_desc.ddpfPixelFormat.dwZBitMask == 0xffffffff,
             "Got unexpected Z mask 0x%08lx.\n", surface_desc.ddpfPixelFormat.dwZBitMask);
+    ok(!(surface_desc.ddsCaps.dwCaps & DDSCAPS_VIDEOMEMORY),
+            "Got unexpected surface caps %#lx.\n", surface_desc.ddsCaps.dwCaps);
     IDirectDrawSurface7_Release(surface);
 
     refcount = IDirectDraw7_Release(ddraw);
@@ -19265,7 +19115,7 @@
     ret = SetWindowPos(window, 0, monitor_rect->left, monitor_rect->top, 0, 0,
             SWP_NOZORDER | SWP_NOSIZE);
     ok(ret, "SetWindowPos failed, error %lu.\n", GetLastError());
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     flush_events();
     ret = ShowWindow(window, SW_RESTORE);
@@ -19334,7 +19184,7 @@
     struct find_different_mode_param param;
     DDSURFACEDESC2 surface_desc;
     IDirectDraw7 *ddraw;
-    DEVMODEA devmode;
+    DEVMODEW devmode;
     HWND window;
     HRESULT hr;
     BOOL ret;
@@ -19351,8 +19201,8 @@
 
     memset(&devmode, 0, sizeof(devmode));
     devmode.dmSize = sizeof(devmode);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
 
     surface_desc.dwSize = sizeof(surface_desc);
     hr = IDirectDraw7_GetDisplayMode(ddraw, &surface_desc);
@@ -19951,16 +19801,11 @@
     DestroyWindow(window);
 }
 
-static unsigned int enum_devices_index;
-
 static HRESULT WINAPI test_enum_devices_caps_callback(char *device_desc, char *device_name,
         D3DDEVICEDESC7 *device_desc7, void *ctx)
 {
     if (IsEqualGUID(&device_desc7->deviceGUID, &IID_IDirect3DTnLHalDevice))
     {
-        ok(enum_devices_index == 2, "Expected index %u.\n", enum_devices_index);
-        todo_wine ok(!strcmp(device_name, "Direct3D T&L HAL"), "Got name %s.\n", debugstr_a(device_name));
-
         ok(device_desc7->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT,
            "TnLHal Device device caps does not have D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
         ok(device_desc7->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX,
@@ -19968,9 +19813,6 @@
     }
     else if (IsEqualGUID(&device_desc7->deviceGUID, &IID_IDirect3DHALDevice))
     {
-        ok(enum_devices_index == 1, "Expected index %u.\n", enum_devices_index);
-        ok(!strcmp(device_name, "Direct3D HAL"), "Got name %s.\n", debugstr_a(device_name));
-
         ok((device_desc7->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
            "HAL Device device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
         ok(device_desc7->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX,
@@ -19978,9 +19820,6 @@
     }
     else if (IsEqualGUID(&device_desc7->deviceGUID, &IID_IDirect3DRGBDevice))
     {
-        ok(enum_devices_index == 0, "Expected index %u.\n", enum_devices_index);
-        ok(!strcmp(device_name, "RGB Emulation"), "Got name %s.\n", debugstr_a(device_name));
-
         ok((device_desc7->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
            "RGB Device device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
         ok((device_desc7->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
@@ -19993,7 +19832,6 @@
         ok(FALSE, "Unexpected device enumerated: \"%s\" \"%s\"\n", device_desc, device_name);
     }
 
-    ++enum_devices_index;
     return DDENUMRET_OK;
 }
 
@@ -20637,292 +20475,11 @@
     DestroyWindow(window);
 }
 
-static void test_yuv_blit(void)
-{
-    DDSURFACEDESC2 surface_desc = {sizeof(surface_desc)};
-    IDirectDrawSurface7 *rgb_surface, *yuv_surface;
-    const unsigned int width = 32, height = 32;
-    uint8_t *buf, *chroma_buf, *u_buf, *v_buf;
-    DDBLTFX fx = {.dwSize = sizeof(fx)};
-    unsigned int color, refcount;
-    IDirectDraw7 *ddraw;
-    HWND window;
-    HRESULT hr;
-
-    static const struct
-    {
-        uint8_t y, u, v;
-        uint32_t rgb_full, rgb_reduced;
-    }
-    tests[] =
-    {
-        {0x10, 0x80, 0x80, 0x000000, 0x101010},
-        {0xeb, 0x80, 0x80, 0xffffff, 0xebebeb},
-        {0x51, 0x5a, 0xf0, 0xff0000, 0xee0e0e},
-        {0x91, 0x36, 0x22, 0x00ff01, 0x0dee0e},
-        {0x29, 0xf0, 0x6e, 0x0000ff, 0x100fef},
-        {0x7e, 0x80, 0x80, 0x808080, 0x7e7e7e},
-        {0x00, 0x80, 0x80, 0x000000, 0x000000},
-        {0xff, 0x80, 0x80, 0xffffff, 0xffffff},
-        {0x00, 0x00, 0x00, 0x008800, 0x008800},
-        {0xff, 0x00, 0x00, 0x4aff14, 0x4cff1c},
-        {0x00, 0xff, 0x00, 0x0024ee, 0x0030e1},
-        {0x00, 0x00, 0xff, 0xb80000, 0xb20000},
-        {0xff, 0xff, 0x00, 0x4affff, 0x4cffff},
-        {0xff, 0x00, 0xff, 0xffe114, 0xffd01c},
-        {0x00, 0xff, 0xff, 0xb800ee, 0xb200e1},
-        {0xff, 0xff, 0xff, 0xff7dff, 0xff78ff},
-    };
-
-    static const struct
-    {
-        DWORD fourcc;
-        const char *str;
-    }
-    formats[] =
-    {
-        {MAKEFOURCC('U','Y','V','Y'), "UYVY"},
-        {MAKEFOURCC('Y','U','Y','2'), "YUY2"},
-        {MAKEFOURCC('Y','V','1','2'), "YV12"},
-        {MAKEFOURCC('N','V','1','2'), "NV12"},
-    };
-
-    window = create_window();
-    ddraw = create_ddraw();
-    hr = IDirectDraw7_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    surface_desc.dwFlags = DDSD_WIDTH | DDSD_HEIGHT | DDSD_CAPS | DDSD_PIXELFORMAT;
-    surface_desc.ddsCaps.dwCaps = DDSCAPS_OFFSCREENPLAIN;
-    surface_desc.dwWidth = 640;
-    surface_desc.dwHeight = 480;
-    init_format_b8g8r8x8(&surface_desc.ddpfPixelFormat);
-
-    hr = IDirectDraw7_CreateSurface(ddraw, &surface_desc, &rgb_surface, NULL);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    for (unsigned int fmt = 0; fmt < ARRAY_SIZE(formats); fmt++)
-    {
-        DWORD format = formats[fmt].fourcc;
-
-        winetest_push_context("format %s", formats[fmt].str);
-
-        surface_desc.dwWidth = width;
-        surface_desc.dwHeight = height;
-        surface_desc.ddpfPixelFormat.dwFlags = DDPF_FOURCC;
-        surface_desc.ddpfPixelFormat.dwFourCC = format;
-        hr = IDirectDraw7_CreateSurface(ddraw, &surface_desc, &yuv_surface, NULL);
-        if (hr != S_OK)
-        {
-            skip("Failed to create surface, hr %#lx.\n", hr);
-            winetest_pop_context();
-            continue;
-        }
-
-        for (unsigned int i = 0; i < ARRAY_SIZE(tests); i++)
-        {
-            winetest_push_context("value (%#x,%#x,%#x)", tests[i].y, tests[i].u, tests[i].v);
-
-            hr = IDirectDrawSurface7_Lock(yuv_surface, NULL, &surface_desc, DDLOCK_WAIT, NULL);
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-            buf = surface_desc.lpSurface;
-            chroma_buf = buf + surface_desc.lPitch * height;
-            if (format == MAKEFOURCC('Y','V','1','2'))
-            {
-                v_buf = chroma_buf;
-                u_buf = chroma_buf + height / 2 * surface_desc.lPitch / 2;
-            }
-            for (unsigned int y = 0; y < height; y++)
-            {
-                for (unsigned int x = 0; x < width; x += 2)
-                {
-                    uint8_t Y = tests[i].y, U = tests[i].u, V = tests[i].v;
-                    if (x < width / 2 && y < height / 2)
-                        Y = U = V = 0x40;
-
-                    if (format == MAKEFOURCC('U','Y','V','Y'))
-                    {
-                        buf[y * surface_desc.lPitch + 2 * x + 0] = U;
-                        buf[y * surface_desc.lPitch + 2 * x + 1] = Y;
-                        buf[y * surface_desc.lPitch + 2 * x + 2] = V;
-                        buf[y * surface_desc.lPitch + 2 * x + 3] = Y;
-                    }
-                    else if (format == MAKEFOURCC('Y','U','Y','2'))
-                    {
-                        buf[y * surface_desc.lPitch + 2 * x + 0] = Y;
-                        buf[y * surface_desc.lPitch + 2 * x + 1] = U;
-                        buf[y * surface_desc.lPitch + 2 * x + 2] = Y;
-                        buf[y * surface_desc.lPitch + 2 * x + 3] = V;
-                    }
-                    else if (format == MAKEFOURCC('Y','V','1','2'))
-                    {
-                        buf[y * surface_desc.lPitch + x + 0] = Y;
-                        buf[y * surface_desc.lPitch + x + 1] = Y;
-                        u_buf[(y / 2) * (surface_desc.lPitch / 2) + (x / 2)] = U;
-                        v_buf[(y / 2) * (surface_desc.lPitch / 2) + (x / 2)] = V;
-                    }
-                    else if (format == MAKEFOURCC('N','V','1','2'))
-                    {
-                        buf[y * surface_desc.lPitch + x + 0] = Y;
-                        buf[y * surface_desc.lPitch + x + 1] = Y;
-                        chroma_buf[(y / 2) * surface_desc.lPitch + 2 * (x / 2) + 0] = U;
-                        chroma_buf[(y / 2) * surface_desc.lPitch + 2 * (x / 2) + 1] = V;
-                    }
-                }
-            }
-            hr = IDirectDrawSurface7_Unlock(yuv_surface, NULL);
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-
-            hr = IDirectDrawSurface7_Blt(rgb_surface, NULL, yuv_surface, NULL, DDBLT_WAIT, NULL);
-            if (hr != D3D_OK)
-            {
-                winetest_pop_context();
-                skip("Failed to blit, hr %#lx.\n", hr);
-                break;
-            }
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-
-            hr = IDirectDrawSurface7_Lock(rgb_surface, NULL, &surface_desc, DDLOCK_WAIT, NULL);
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-
-            for (unsigned int y = 0; y < 4; y++)
-            {
-                for (unsigned int x = 0; x < 4; x++)
-                {
-                    unsigned int xcoord = (1 + 2 * x) * 640 / 8;
-                    unsigned int ycoord = (1 + 2 * y) * 480 / 8;
-
-                    color = ((uint32_t *)((uint8_t *)surface_desc.lpSurface + ycoord * surface_desc.lPitch))[xcoord];
-                    color &= 0xffffff;
-
-                    if (x < 2 && y < 2)
-                        ok(compare_color(color, 0x008400, 1),
-                                "Got color %#x at (%u, %u).\n", color, xcoord, ycoord);
-                    else
-                        ok(compare_color(color, tests[i].rgb_full, 1)
-                                || compare_color(color, tests[i].rgb_reduced, 1),
-                                "Got color %#x at (%u, %u), expected %#x.\n", color, xcoord, ycoord, tests[i].rgb_full);
-                }
-            }
-            hr = IDirectDrawSurface7_Unlock(rgb_surface, NULL);
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-
-            winetest_pop_context();
-        }
-
-        IDirectDrawSurface7_Release(yuv_surface);
-        winetest_pop_context();
-    }
-
-    IDirectDrawSurface7_Release(rgb_surface);
-    refcount = IDirectDraw7_Release(ddraw);
-    ok(!refcount, "Device has %u references left.\n", refcount);
-    DestroyWindow(window);
-}
-
-static void test_blit_to_self(void)
-{
-    DDSURFACEDESC2 surface_desc = {sizeof(surface_desc)};
-    const unsigned int width = 64, height = 64;
-    IDirectDrawSurface7 *surface;
-    unsigned int refcount;
-    IDirectDraw7 *ddraw;
-    HWND window;
-    HRESULT hr;
-
-    struct
-    {
-        RECT src, dst;
-    }
-    tests[] =
-    {
-        {{16, 16, 24, 24}, {18, 18, 26, 26}},
-        {{16, 16, 24, 24}, {14, 14, 22, 22}},
-        {{16, 16, 24, 24}, {12, 12, 28, 28}},
-        {{16, 16, 24, 24}, {18, 18, 22, 22}},
-    };
-
-    window = create_window();
-    ddraw = create_ddraw();
-    hr = IDirectDraw7_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    surface_desc.dwFlags = DDSD_WIDTH | DDSD_HEIGHT | DDSD_CAPS | DDSD_PIXELFORMAT;
-    surface_desc.ddsCaps.dwCaps = DDSCAPS_OFFSCREENPLAIN | DDSCAPS_SYSTEMMEMORY;
-    surface_desc.dwWidth = width;
-    surface_desc.dwHeight = height;
-    init_format_b8g8r8x8(&surface_desc.ddpfPixelFormat);
-
-    hr = IDirectDraw7_CreateSurface(ddraw, &surface_desc, &surface, NULL);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    for (unsigned int i = 0; i < ARRAY_SIZE(tests); ++i)
-    {
-        hr = IDirectDrawSurface7_Lock(surface, NULL, &surface_desc, DDLOCK_WAIT, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-        for (unsigned int y = 0; y < height; ++y)
-        {
-            for (unsigned int x = 0; x < width; ++x)
-                ((unsigned int *)surface_desc.lpSurface)[y * width + x] = (y << 8) | x;
-        }
-        hr = IDirectDrawSurface7_Unlock(surface, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-        hr = IDirectDrawSurface7_Blt(surface, &tests[i].dst, surface, &tests[i].src, DDBLT_WAIT, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-        hr = IDirectDrawSurface7_Lock(surface, NULL, &surface_desc, DDLOCK_WAIT | DDLOCK_READONLY, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-        for (unsigned int y = 0; y < height; ++y)
-        {
-            for (unsigned int x = 0; x < width; ++x)
-            {
-                unsigned int colour = ((unsigned int *)surface_desc.lpSurface)[y * width + x];
-
-                if (x >= tests[i].dst.left && x < tests[i].dst.right
-                        && y >= tests[i].dst.top && y < tests[i].dst.bottom)
-                {
-                    unsigned int src_x = (x - tests[i].dst.left)
-                            * (tests[i].src.right - tests[i].src.left)
-                            / (tests[i].dst.right - tests[i].dst.left)
-                            + tests[i].src.left;
-                    unsigned int src_y = (y - tests[i].dst.top)
-                            * (tests[i].src.bottom - tests[i].src.top)
-                            / (tests[i].dst.bottom - tests[i].dst.top)
-                            + tests[i].src.top;
-                    if (colour != ((src_y << 8) | src_x))
-                    {
-                        ok(0, "Got colour 0x%08x at (%u, %u), expected 0x%08x.\n",
-                                colour, x, y, ((src_y << 8) | src_x));
-                        goto out;
-                    }
-                }
-                else
-                {
-                    if (colour != ((y << 8) | x))
-                    {
-                        ok(0, "Got colour 0x%08x at (%u, %u).\n", colour, x, y);
-                        goto out;
-                    }
-                }
-            }
-        }
-out:
-        hr = IDirectDrawSurface7_Unlock(surface, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    }
-
-    IDirectDrawSurface7_Release(surface);
-    refcount = IDirectDraw7_Release(ddraw);
-    ok(!refcount, "Device has %u references left.\n", refcount);
-    DestroyWindow(window);
-}
-
 START_TEST(ddraw7)
 {
     DDDEVICEIDENTIFIER2 identifier;
     HMODULE module, dwmapi;
-    DEVMODEA current_mode;
+    DEVMODEW current_mode;
     IDirectDraw7 *ddraw;
     IDirect3D7 *d3d;
     BOOL hal_ok;
@@ -20967,9 +20524,9 @@
 
     memset(&current_mode, 0, sizeof(current_mode));
     current_mode.dmSize = sizeof(current_mode);
-    ok(EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &current_mode), "Failed to get display mode.\n");
+    ok(EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &current_mode), "Failed to get display mode.\n");
     registry_mode.dmSize = sizeof(registry_mode);
-    ok(EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &registry_mode), "Failed to get display mode.\n");
+    ok(EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &registry_mode), "Failed to get display mode.\n");
     if (registry_mode.dmPelsWidth != current_mode.dmPelsWidth
             || registry_mode.dmPelsHeight != current_mode.dmPelsHeight)
     {
@@ -20979,8 +20536,6 @@
 
     if ((dwmapi = LoadLibraryA("dwmapi.dll")))
         pDwmIsCompositionEnabled = (void *)GetProcAddress(dwmapi, "DwmIsCompositionEnabled");
-
-    start_foreground_window_thread();
 
     test_process_vertices();
     test_coop_level_create_device_window();
@@ -21099,6 +20654,4 @@
     test_vb_desc();
     test_d3d_state_reset();
     test_sysmem_x_channel();
-    test_yuv_blit();
-    test_blit_to_self();
 }
