diff --git a/dlls/opengl32/unixlib.h b/dlls/opengl32/unixlib.h
--- a/dlls/opengl32/unixlib.h
+++ b/dlls/opengl32/unixlib.h
@@ -2658,13 +2658,6 @@
     GLuint arg3;
     GLuint arg3Rep;
     GLuint arg3Mod;
-};
-
-struct glAlphaFuncx_params
-{
-    TEB *teb;
-    GLenum func;
-    GLfixed ref;
 };
 
 struct glAlphaFuncxOES_params
@@ -3598,38 +3591,6 @@
     GLenum filter;
 };
 
-struct glBlitFramebufferLayerEXT_params
-{
-    TEB *teb;
-    GLint srcX0;
-    GLint srcY0;
-    GLint srcX1;
-    GLint srcY1;
-    GLint srcLayer;
-    GLint dstX0;
-    GLint dstY0;
-    GLint dstX1;
-    GLint dstY1;
-    GLint dstLayer;
-    GLbitfield mask;
-    GLenum filter;
-};
-
-struct glBlitFramebufferLayersEXT_params
-{
-    TEB *teb;
-    GLint srcX0;
-    GLint srcY0;
-    GLint srcX1;
-    GLint srcY1;
-    GLint dstX0;
-    GLint dstY0;
-    GLint dstX1;
-    GLint dstY1;
-    GLbitfield mask;
-    GLenum filter;
-};
-
 struct glBlitNamedFramebuffer_params
 {
     TEB *teb;
@@ -3691,17 +3652,6 @@
     GLboolean commit;
 };
 
-struct glBufferPageCommitmentMemNV_params
-{
-    TEB *teb;
-    GLenum target;
-    GLintptr offset;
-    GLsizeiptr size;
-    GLuint memory;
-    GLuint64 memOffset;
-    GLboolean commit;
-};
-
 struct glBufferParameteriAPPLE_params
 {
     TEB *teb;
@@ -3894,7 +3844,7 @@
     GLuint alpha;
 };
 
-struct glClearColorx_params
+struct glClearColorxOES_params
 {
     TEB *teb;
     GLfixed red;
@@ -3903,15 +3853,6 @@
     GLfixed alpha;
 };
 
-struct glClearColorxOES_params
-{
-    TEB *teb;
-    GLfixed red;
-    GLfixed green;
-    GLfixed blue;
-    GLfixed alpha;
-};
-
 struct glClearDepthdNV_params
 {
     TEB *teb;
@@ -3928,12 +3869,6 @@
 {
     TEB *teb;
     GLclampf depth;
-};
-
-struct glClearDepthx_params
-{
-    TEB *teb;
-    GLfixed depth;
 };
 
 struct glClearDepthxOES_params
@@ -4097,25 +4032,11 @@
     GLenum depth;
 };
 
-struct glClipPlanef_params
-{
-    TEB *teb;
-    GLenum p;
-    const GLfloat *eqn;
-};
-
 struct glClipPlanefOES_params
 {
     TEB *teb;
     GLenum plane;
     const GLfloat *equation;
-};
-
-struct glClipPlanex_params
-{
-    TEB *teb;
-    GLenum plane;
-    const GLfixed *equation;
 };
 
 struct glClipPlanexOES_params
@@ -4244,15 +4165,6 @@
     TEB *teb;
     const GLubyte *c;
     const GLfloat *v;
-};
-
-struct glColor4x_params
-{
-    TEB *teb;
-    GLfixed red;
-    GLfixed green;
-    GLfixed blue;
-    GLfixed alpha;
 };
 
 struct glColor4xOES_params
@@ -5633,13 +5545,6 @@
     GLuint *samplers;
 };
 
-struct glCreateSemaphoresNV_params
-{
-    TEB *teb;
-    GLsizei n;
-    GLuint *semaphores;
-};
-
 struct glCreateShader_params
 {
     TEB *teb;
@@ -6190,13 +6095,6 @@
     GLclampf f;
 };
 
-struct glDepthRangex_params
-{
-    TEB *teb;
-    GLfixed n;
-    GLfixed f;
-};
-
 struct glDepthRangexOES_params
 {
     TEB *teb;
@@ -6554,20 +6452,6 @@
     GLint first;
     GLsizei count;
     GLsizei width;
-};
-
-struct glDrawMeshTasksEXT_params
-{
-    TEB *teb;
-    GLuint num_groups_x;
-    GLuint num_groups_y;
-    GLuint num_groups_z;
-};
-
-struct glDrawMeshTasksIndirectEXT_params
-{
-    TEB *teb;
-    GLintptr indirect;
 };
 
 struct glDrawMeshTasksIndirectNV_params
@@ -7184,25 +7068,11 @@
     const GLfloat *points;
 };
 
-struct glFogx_params
+struct glFogxOES_params
 {
     TEB *teb;
     GLenum pname;
     GLfixed param;
-};
-
-struct glFogxOES_params
-{
-    TEB *teb;
-    GLenum pname;
-    GLfixed param;
-};
-
-struct glFogxv_params
-{
-    TEB *teb;
-    GLenum pname;
-    const GLfixed *param;
 };
 
 struct glFogxvOES_params
@@ -7572,7 +7442,7 @@
     GLuint buffer;
 };
 
-struct glFrustumf_params
+struct glFrustumfOES_params
 {
     TEB *teb;
     GLfloat l;
@@ -7583,18 +7453,7 @@
     GLfloat f;
 };
 
-struct glFrustumfOES_params
-{
-    TEB *teb;
-    GLfloat l;
-    GLfloat r;
-    GLfloat b;
-    GLfloat t;
-    GLfloat n;
-    GLfloat f;
-};
-
-struct glFrustumx_params
+struct glFrustumxOES_params
 {
     TEB *teb;
     GLfixed l;
@@ -7605,17 +7464,6 @@
     GLfixed f;
 };
 
-struct glFrustumxOES_params
-{
-    TEB *teb;
-    GLfixed l;
-    GLfixed r;
-    GLfixed b;
-    GLfixed t;
-    GLfixed n;
-    GLfixed f;
-};
-
 struct glGenAsyncMarkersSGIX_params
 {
     TEB *teb;
@@ -8127,25 +7975,11 @@
     void *data;
 };
 
-struct glGetClipPlanef_params
+struct glGetClipPlanefOES_params
 {
     TEB *teb;
     GLenum plane;
     GLfloat *equation;
-};
-
-struct glGetClipPlanefOES_params
-{
-    TEB *teb;
-    GLenum plane;
-    GLfloat *equation;
-};
-
-struct glGetClipPlanex_params
-{
-    TEB *teb;
-    GLenum plane;
-    GLfixed *equation;
 };
 
 struct glGetClipPlanexOES_params
@@ -8427,7 +8261,7 @@
     GLuint count;
     GLsizei bufSize;
     GLenum *categories;
-    GLenum *severities;
+    GLuint *severities;
     GLuint *ids;
     GLsizei *lengths;
     GLchar *message;
@@ -8509,13 +8343,6 @@
     GLuint *queryId;
 };
 
-struct glGetFixedv_params
-{
-    TEB *teb;
-    GLenum pname;
-    GLfixed *params;
-};
-
 struct glGetFixedvOES_params
 {
     TEB *teb;
@@ -8903,14 +8730,6 @@
     GLfixed *params;
 };
 
-struct glGetLightxv_params
-{
-    TEB *teb;
-    GLenum light;
-    GLenum pname;
-    GLfixed *params;
-};
-
 struct glGetListParameterfvSGIX_params
 {
     TEB *teb;
@@ -9011,14 +8830,6 @@
     GLenum face;
     GLenum pname;
     GLfixed param;
-};
-
-struct glGetMaterialxv_params
-{
-    TEB *teb;
-    GLenum face;
-    GLenum pname;
-    GLfixed *params;
 };
 
 struct glGetMemoryObjectDetachedResourcesuivNV_params
@@ -10227,14 +10038,6 @@
     GLint *params;
 };
 
-struct glGetSemaphoreParameterivNV_params
-{
-    TEB *teb;
-    GLuint semaphore;
-    GLenum pname;
-    GLint *params;
-};
-
 struct glGetSemaphoreParameterui64vEXT_params
 {
     TEB *teb;
@@ -10390,14 +10193,6 @@
     GLint *param;
 };
 
-struct glGetTexEnvxv_params
-{
-    TEB *teb;
-    GLenum target;
-    GLenum pname;
-    GLfixed *params;
-};
-
 struct glGetTexEnvxvOES_params
 {
     TEB *teb;
@@ -10469,14 +10264,6 @@
     GLenum target;
     GLenum pname;
     void **params;
-};
-
-struct glGetTexParameterxv_params
-{
-    TEB *teb;
-    GLenum target;
-    GLenum pname;
-    GLfixed *params;
 };
 
 struct glGetTexParameterxvOES_params
@@ -12319,70 +12106,34 @@
     GLint param;
 };
 
-struct glLightModelx_params
+struct glLightModelxOES_params
 {
     TEB *teb;
     GLenum pname;
     GLfixed param;
 };
 
-struct glLightModelxOES_params
-{
-    TEB *teb;
+struct glLightModelxvOES_params
+{
+    TEB *teb;
+    GLenum pname;
+    const GLfixed *param;
+};
+
+struct glLightxOES_params
+{
+    TEB *teb;
+    GLenum light;
     GLenum pname;
     GLfixed param;
 };
 
-struct glLightModelxv_params
-{
-    TEB *teb;
-    GLenum pname;
-    const GLfixed *param;
-};
-
-struct glLightModelxvOES_params
-{
-    TEB *teb;
-    GLenum pname;
-    const GLfixed *param;
-};
-
-struct glLightx_params
+struct glLightxvOES_params
 {
     TEB *teb;
     GLenum light;
     GLenum pname;
-    GLfixed param;
-};
-
-struct glLightxOES_params
-{
-    TEB *teb;
-    GLenum light;
-    GLenum pname;
-    GLfixed param;
-};
-
-struct glLightxv_params
-{
-    TEB *teb;
-    GLenum light;
-    GLenum pname;
     const GLfixed *params;
-};
-
-struct glLightxvOES_params
-{
-    TEB *teb;
-    GLenum light;
-    GLenum pname;
-    const GLfixed *params;
-};
-
-struct glLineWidthx_params
-{
-    TEB *teb;
-    GLfixed width;
 };
 
 struct glLineWidthxOES_params
@@ -12451,12 +12202,6 @@
 {
     TEB *teb;
     GLbitfield mask;
-};
-
-struct glLoadMatrixx_params
-{
-    TEB *teb;
-    const GLfixed *m;
 };
 
 struct glLoadMatrixxOES_params
@@ -12812,28 +12557,12 @@
     const GLfloat *points;
 };
 
-struct glMaterialx_params
+struct glMaterialxOES_params
 {
     TEB *teb;
     GLenum face;
     GLenum pname;
     GLfixed param;
-};
-
-struct glMaterialxOES_params
-{
-    TEB *teb;
-    GLenum face;
-    GLenum pname;
-    GLfixed param;
-};
-
-struct glMaterialxv_params
-{
-    TEB *teb;
-    GLenum face;
-    GLenum pname;
-    const GLfixed *param;
 };
 
 struct glMaterialxvOES_params
@@ -13136,12 +12865,6 @@
     GLboolean sink;
 };
 
-struct glMultMatrixx_params
-{
-    TEB *teb;
-    const GLfixed *m;
-};
-
 struct glMultMatrixxOES_params
 {
     TEB *teb;
@@ -13360,29 +13083,12 @@
     GLsizei stride;
 };
 
-struct glMultiDrawMeshTasksIndirectCountEXT_params
+struct glMultiDrawMeshTasksIndirectCountNV_params
 {
     TEB *teb;
     GLintptr indirect;
     GLintptr drawcount;
     GLsizei maxdrawcount;
-    GLsizei stride;
-};
-
-struct glMultiDrawMeshTasksIndirectCountNV_params
-{
-    TEB *teb;
-    GLintptr indirect;
-    GLintptr drawcount;
-    GLsizei maxdrawcount;
-    GLsizei stride;
-};
-
-struct glMultiDrawMeshTasksIndirectEXT_params
-{
-    TEB *teb;
-    GLintptr indirect;
-    GLsizei drawcount;
     GLsizei stride;
 };
 
@@ -14346,16 +14052,6 @@
     TEB *teb;
     GLenum target;
     GLshort * v;
-};
-
-struct glMultiTexCoord4x_params
-{
-    TEB *teb;
-    GLenum texture;
-    GLfixed s;
-    GLfixed t;
-    GLfixed r;
-    GLfixed q;
 };
 
 struct glMultiTexCoord4xOES_params
@@ -14891,17 +14587,6 @@
     GLboolean commit;
 };
 
-struct glNamedBufferPageCommitmentMemNV_params
-{
-    TEB *teb;
-    GLuint buffer;
-    GLintptr offset;
-    GLsizeiptr size;
-    GLuint memory;
-    GLuint64 memOffset;
-    GLboolean commit;
-};
-
 struct glNamedBufferStorage_params
 {
     TEB *teb;
@@ -15129,17 +14814,6 @@
     GLint layer;
 };
 
-struct glNamedFramebufferTextureMultiviewOVR_params
-{
-    TEB *teb;
-    GLuint framebuffer;
-    GLenum attachment;
-    GLuint texture;
-    GLint level;
-    GLint baseViewIndex;
-    GLsizei numViews;
-};
-
 struct glNamedProgramLocalParameter4dEXT_params
 {
     TEB *teb;
@@ -15382,14 +15056,6 @@
     const GLhalfNV *v;
 };
 
-struct glNormal3x_params
-{
-    TEB *teb;
-    GLfixed nx;
-    GLfixed ny;
-    GLfixed nz;
-};
-
 struct glNormal3xOES_params
 {
     TEB *teb;
@@ -15565,7 +15231,7 @@
     GLenum ret;
 };
 
-struct glOrthof_params
+struct glOrthofOES_params
 {
     TEB *teb;
     GLfloat l;
@@ -15574,28 +15240,6 @@
     GLfloat t;
     GLfloat n;
     GLfloat f;
-};
-
-struct glOrthofOES_params
-{
-    TEB *teb;
-    GLfloat l;
-    GLfloat r;
-    GLfloat b;
-    GLfloat t;
-    GLfloat n;
-    GLfloat f;
-};
-
-struct glOrthox_params
-{
-    TEB *teb;
-    GLfixed l;
-    GLfixed r;
-    GLfixed b;
-    GLfixed t;
-    GLfixed n;
-    GLfixed f;
 };
 
 struct glOrthoxOES_params
@@ -15722,7 +15366,7 @@
     GLbitfield fontStyle;
     GLuint pathParameterTemplate;
     GLfloat emScale;
-    GLuint *baseAndCount;
+    GLuint baseAndCount[2];
     GLenum ret;
 };
 
@@ -16069,31 +15713,11 @@
     const GLint *params;
 };
 
-struct glPointParameterx_params
-{
-    TEB *teb;
-    GLenum pname;
-    GLfixed param;
-};
-
-struct glPointParameterxv_params
+struct glPointParameterxvOES_params
 {
     TEB *teb;
     GLenum pname;
     const GLfixed *params;
-};
-
-struct glPointParameterxvOES_params
-{
-    TEB *teb;
-    GLenum pname;
-    const GLfixed *params;
-};
-
-struct glPointSizex_params
-{
-    TEB *teb;
-    GLfixed size;
 };
 
 struct glPointSizexOES_params
@@ -16137,13 +15761,6 @@
     TEB *teb;
     GLfloat factor;
     GLfloat bias;
-};
-
-struct glPolygonOffsetx_params
-{
-    TEB *teb;
-    GLfixed factor;
-    GLfixed units;
 };
 
 struct glPolygonOffsetxOES_params
@@ -18484,7 +18101,7 @@
     TEB *teb;
 };
 
-struct glRotatex_params
+struct glRotatexOES_params
 {
     TEB *teb;
     GLfixed angle;
@@ -18493,15 +18110,6 @@
     GLfixed z;
 };
 
-struct glRotatexOES_params
-{
-    TEB *teb;
-    GLfixed angle;
-    GLfixed x;
-    GLfixed y;
-    GLfixed z;
-};
-
 struct glSampleCoverage_params
 {
     TEB *teb;
@@ -18513,13 +18121,6 @@
 {
     TEB *teb;
     GLfloat value;
-    GLboolean invert;
-};
-
-struct glSampleCoveragex_params
-{
-    TEB *teb;
-    GLclampx value;
     GLboolean invert;
 };
 
@@ -18617,14 +18218,6 @@
     GLuint sampler;
     GLenum pname;
     const GLint *param;
-};
-
-struct glScalex_params
-{
-    TEB *teb;
-    GLfixed x;
-    GLfixed y;
-    GLfixed z;
 };
 
 struct glScalexOES_params
@@ -18985,14 +18578,6 @@
 {
     TEB *teb;
     GLenum target;
-};
-
-struct glSemaphoreParameterivNV_params
-{
-    TEB *teb;
-    GLuint semaphore;
-    GLenum pname;
-    const GLint *params;
 };
 
 struct glSemaphoreParameterui64vEXT_params
@@ -19078,7 +18663,7 @@
     TEB *teb;
     GLsizei count;
     const GLuint *shaders;
-    GLenum binaryFormat;
+    GLenum binaryformat;
     const void *binary;
     GLsizei length;
 };
@@ -20036,28 +19621,12 @@
     const void **pointer;
 };
 
-struct glTexEnvx_params
+struct glTexEnvxOES_params
 {
     TEB *teb;
     GLenum target;
     GLenum pname;
     GLfixed param;
-};
-
-struct glTexEnvxOES_params
-{
-    TEB *teb;
-    GLenum target;
-    GLenum pname;
-    GLfixed param;
-};
-
-struct glTexEnvxv_params
-{
-    TEB *teb;
-    GLenum target;
-    GLenum pname;
-    const GLfixed *params;
 };
 
 struct glTexEnvxvOES_params
@@ -20201,23 +19770,6 @@
     GLboolean commit;
 };
 
-struct glTexPageCommitmentMemNV_params
-{
-    TEB *teb;
-    GLenum target;
-    GLint layer;
-    GLint level;
-    GLint xoffset;
-    GLint yoffset;
-    GLint zoffset;
-    GLsizei width;
-    GLsizei height;
-    GLsizei depth;
-    GLuint memory;
-    GLuint64 offset;
-    GLboolean commit;
-};
-
 struct glTexParameterIiv_params
 {
     TEB *teb;
@@ -20250,28 +19802,12 @@
     const GLuint *params;
 };
 
-struct glTexParameterx_params
+struct glTexParameterxOES_params
 {
     TEB *teb;
     GLenum target;
     GLenum pname;
     GLfixed param;
-};
-
-struct glTexParameterxOES_params
-{
-    TEB *teb;
-    GLenum target;
-    GLenum pname;
-    GLfixed param;
-};
-
-struct glTexParameterxv_params
-{
-    TEB *teb;
-    GLenum target;
-    GLenum pname;
-    const GLfixed *params;
 };
 
 struct glTexParameterxvOES_params
@@ -20298,26 +19834,7 @@
     GLsizei width;
 };
 
-struct glTexStorage1DEXT_params
-{
-    TEB *teb;
-    GLenum target;
-    GLsizei levels;
-    GLenum internalformat;
-    GLsizei width;
-};
-
 struct glTexStorage2D_params
-{
-    TEB *teb;
-    GLenum target;
-    GLsizei levels;
-    GLenum internalformat;
-    GLsizei width;
-    GLsizei height;
-};
-
-struct glTexStorage2DEXT_params
 {
     TEB *teb;
     GLenum target;
@@ -20339,17 +19856,6 @@
 };
 
 struct glTexStorage3D_params
-{
-    TEB *teb;
-    GLenum target;
-    GLsizei levels;
-    GLenum internalformat;
-    GLsizei width;
-    GLsizei height;
-    GLsizei depth;
-};
-
-struct glTexStorage3DEXT_params
 {
     TEB *teb;
     GLenum target;
@@ -20718,23 +20224,6 @@
     GLboolean commit;
 };
 
-struct glTexturePageCommitmentMemNV_params
-{
-    TEB *teb;
-    GLuint texture;
-    GLint layer;
-    GLint level;
-    GLint xoffset;
-    GLint yoffset;
-    GLint zoffset;
-    GLsizei width;
-    GLsizei height;
-    GLsizei depth;
-    GLuint memory;
-    GLuint64 offset;
-    GLboolean commit;
-};
-
 struct glTextureParameterIiv_params
 {
     TEB *teb;
@@ -21221,14 +20710,6 @@
     const GLfloat *transformValues;
 };
 
-struct glTranslatex_params
-{
-    TEB *teb;
-    GLfixed x;
-    GLfixed y;
-    GLfixed z;
-};
-
 struct glTranslatexOES_params
 {
     TEB *teb;
@@ -25697,6 +25178,12 @@
 {
     TEB *teb;
     void *pointer;
+};
+
+struct wglGetCurrentReadDCARB_params
+{
+    TEB *teb;
+    HDC ret;
 };
 
 struct wglGetExtensionsStringARB_params
@@ -26212,7 +25699,6 @@
     unix_glAlphaFragmentOp1ATI,
     unix_glAlphaFragmentOp2ATI,
     unix_glAlphaFragmentOp3ATI,
-    unix_glAlphaFuncx,
     unix_glAlphaFuncxOES,
     unix_glAlphaToCoverageDitherControlNV,
     unix_glApplyFramebufferAttachmentCMAAINTEL,
@@ -26329,15 +25815,12 @@
     unix_glBlendParameteriNV,
     unix_glBlitFramebuffer,
     unix_glBlitFramebufferEXT,
-    unix_glBlitFramebufferLayerEXT,
-    unix_glBlitFramebufferLayersEXT,
     unix_glBlitNamedFramebuffer,
     unix_glBufferAddressRangeNV,
     unix_glBufferAttachMemoryNV,
     unix_glBufferData,
     unix_glBufferDataARB,
     unix_glBufferPageCommitmentARB,
-    unix_glBufferPageCommitmentMemNV,
     unix_glBufferParameteriAPPLE,
     unix_glBufferRegionEnabled,
     unix_glBufferStorage,
@@ -26361,12 +25844,10 @@
     unix_glClearBufferuiv,
     unix_glClearColorIiEXT,
     unix_glClearColorIuiEXT,
-    unix_glClearColorx,
     unix_glClearColorxOES,
     unix_glClearDepthdNV,
     unix_glClearDepthf,
     unix_glClearDepthfOES,
-    unix_glClearDepthx,
     unix_glClearDepthxOES,
     unix_glClearNamedBufferData,
     unix_glClearNamedBufferDataEXT,
@@ -26385,9 +25866,7 @@
     unix_glClientWaitSemaphoreui64NVX,
     unix_glClientWaitSync,
     unix_glClipControl,
-    unix_glClipPlanef,
     unix_glClipPlanefOES,
-    unix_glClipPlanex,
     unix_glClipPlanexOES,
     unix_glColor3fVertex3fSUN,
     unix_glColor3fVertex3fvSUN,
@@ -26403,7 +25882,6 @@
     unix_glColor4ubVertex2fvSUN,
     unix_glColor4ubVertex3fSUN,
     unix_glColor4ubVertex3fvSUN,
-    unix_glColor4x,
     unix_glColor4xOES,
     unix_glColor4xvOES,
     unix_glColorFormatNV,
@@ -26533,7 +26011,6 @@
     unix_glCreateQueries,
     unix_glCreateRenderbuffers,
     unix_glCreateSamplers,
-    unix_glCreateSemaphoresNV,
     unix_glCreateShader,
     unix_glCreateShaderObjectARB,
     unix_glCreateShaderProgramEXT,
@@ -26606,7 +26083,6 @@
     unix_glDepthRangedNV,
     unix_glDepthRangef,
     unix_glDepthRangefOES,
-    unix_glDepthRangex,
     unix_glDepthRangexOES,
     unix_glDetachObjectARB,
     unix_glDetachShader,
@@ -26650,8 +26126,6 @@
     unix_glDrawElementsInstancedBaseVertexBaseInstance,
     unix_glDrawElementsInstancedEXT,
     unix_glDrawMeshArraysSUN,
-    unix_glDrawMeshTasksEXT,
-    unix_glDrawMeshTasksIndirectEXT,
     unix_glDrawMeshTasksIndirectNV,
     unix_glDrawMeshTasksNV,
     unix_glDrawRangeElementArrayAPPLE,
@@ -26739,9 +26213,7 @@
     unix_glFogCoordhNV,
     unix_glFogCoordhvNV,
     unix_glFogFuncSGIS,
-    unix_glFogx,
     unix_glFogxOES,
-    unix_glFogxv,
     unix_glFogxvOES,
     unix_glFragmentColorMaterialSGIX,
     unix_glFragmentCoverageColorNV,
@@ -26786,9 +26258,7 @@
     unix_glFramebufferTextureLayerEXT,
     unix_glFramebufferTextureMultiviewOVR,
     unix_glFreeObjectBufferATI,
-    unix_glFrustumf,
     unix_glFrustumfOES,
-    unix_glFrustumx,
     unix_glFrustumxOES,
     unix_glGenAsyncMarkersSGIX,
     unix_glGenBuffers,
@@ -26853,9 +26323,7 @@
     unix_glGetBufferPointervARB,
     unix_glGetBufferSubData,
     unix_glGetBufferSubDataARB,
-    unix_glGetClipPlanef,
     unix_glGetClipPlanefOES,
-    unix_glGetClipPlanex,
     unix_glGetClipPlanexOES,
     unix_glGetColorTable,
     unix_glGetColorTableEXT,
@@ -26897,7 +26365,6 @@
     unix_glGetFinalCombinerInputParameterfvNV,
     unix_glGetFinalCombinerInputParameterivNV,
     unix_glGetFirstPerfQueryIdINTEL,
-    unix_glGetFixedv,
     unix_glGetFixedvOES,
     unix_glGetFloatIndexedvEXT,
     unix_glGetFloati_v,
@@ -26945,7 +26412,6 @@
     unix_glGetInvariantFloatvEXT,
     unix_glGetInvariantIntegervEXT,
     unix_glGetLightxOES,
-    unix_glGetLightxv,
     unix_glGetListParameterfvSGIX,
     unix_glGetListParameterivSGIX,
     unix_glGetLocalConstantBooleanvEXT,
@@ -26958,7 +26424,6 @@
     unix_glGetMapParameterivNV,
     unix_glGetMapxvOES,
     unix_glGetMaterialxOES,
-    unix_glGetMaterialxv,
     unix_glGetMemoryObjectDetachedResourcesuivNV,
     unix_glGetMemoryObjectParameterivEXT,
     unix_glGetMinmax,
@@ -27097,7 +26562,6 @@
     unix_glGetSamplerParameterIuiv,
     unix_glGetSamplerParameterfv,
     unix_glGetSamplerParameteriv,
-    unix_glGetSemaphoreParameterivNV,
     unix_glGetSemaphoreParameterui64vEXT,
     unix_glGetSeparableFilter,
     unix_glGetSeparableFilterEXT,
@@ -27116,7 +26580,6 @@
     unix_glGetSynciv,
     unix_glGetTexBumpParameterfvATI,
     unix_glGetTexBumpParameterivATI,
-    unix_glGetTexEnvxv,
     unix_glGetTexEnvxvOES,
     unix_glGetTexFilterFuncSGIS,
     unix_glGetTexGenxvOES,
@@ -27126,7 +26589,6 @@
     unix_glGetTexParameterIuiv,
     unix_glGetTexParameterIuivEXT,
     unix_glGetTexParameterPointervAPPLE,
-    unix_glGetTexParameterxv,
     unix_glGetTexParameterxvOES,
     unix_glGetTextureHandleARB,
     unix_glGetTextureHandleNV,
@@ -27349,15 +26811,10 @@
     unix_glLGPUNamedBufferSubDataNVX,
     unix_glLabelObjectEXT,
     unix_glLightEnviSGIX,
-    unix_glLightModelx,
     unix_glLightModelxOES,
-    unix_glLightModelxv,
     unix_glLightModelxvOES,
-    unix_glLightx,
     unix_glLightxOES,
-    unix_glLightxv,
     unix_glLightxvOES,
-    unix_glLineWidthx,
     unix_glLineWidthxOES,
     unix_glLinkProgram,
     unix_glLinkProgramARB,
@@ -27367,7 +26824,6 @@
     unix_glListParameteriSGIX,
     unix_glListParameterivSGIX,
     unix_glLoadIdentityDeformationMapSGIX,
-    unix_glLoadMatrixx,
     unix_glLoadMatrixxOES,
     unix_glLoadProgramNV,
     unix_glLoadTransposeMatrixd,
@@ -27410,9 +26866,7 @@
     unix_glMapVertexAttrib1fAPPLE,
     unix_glMapVertexAttrib2dAPPLE,
     unix_glMapVertexAttrib2fAPPLE,
-    unix_glMaterialx,
     unix_glMaterialxOES,
-    unix_glMaterialxv,
     unix_glMaterialxvOES,
     unix_glMatrixFrustumEXT,
     unix_glMatrixIndexPointerARB,
@@ -27453,7 +26907,6 @@
     unix_glMinSampleShadingARB,
     unix_glMinmax,
     unix_glMinmaxEXT,
-    unix_glMultMatrixx,
     unix_glMultMatrixxOES,
     unix_glMultTransposeMatrixd,
     unix_glMultTransposeMatrixdARB,
@@ -27478,9 +26931,7 @@
     unix_glMultiDrawElementsIndirectBindlessNV,
     unix_glMultiDrawElementsIndirectCount,
     unix_glMultiDrawElementsIndirectCountARB,
-    unix_glMultiDrawMeshTasksIndirectCountEXT,
     unix_glMultiDrawMeshTasksIndirectCountNV,
-    unix_glMultiDrawMeshTasksIndirectEXT,
     unix_glMultiDrawMeshTasksIndirectNV,
     unix_glMultiDrawRangeElementArrayAPPLE,
     unix_glMultiModeDrawArraysIBM,
@@ -27604,7 +27055,6 @@
     unix_glMultiTexCoord4sv,
     unix_glMultiTexCoord4svARB,
     unix_glMultiTexCoord4svSGIS,
-    unix_glMultiTexCoord4x,
     unix_glMultiTexCoord4xOES,
     unix_glMultiTexCoord4xvOES,
     unix_glMultiTexCoordP1ui,
@@ -27659,7 +27109,6 @@
     unix_glNamedBufferDataEXT,
     unix_glNamedBufferPageCommitmentARB,
     unix_glNamedBufferPageCommitmentEXT,
-    unix_glNamedBufferPageCommitmentMemNV,
     unix_glNamedBufferStorage,
     unix_glNamedBufferStorageEXT,
     unix_glNamedBufferStorageExternalEXT,
@@ -27685,7 +27134,6 @@
     unix_glNamedFramebufferTextureFaceEXT,
     unix_glNamedFramebufferTextureLayer,
     unix_glNamedFramebufferTextureLayerEXT,
-    unix_glNamedFramebufferTextureMultiviewOVR,
     unix_glNamedProgramLocalParameter4dEXT,
     unix_glNamedProgramLocalParameter4dvEXT,
     unix_glNamedProgramLocalParameter4fEXT,
@@ -27711,7 +27159,6 @@
     unix_glNormal3fVertex3fvSUN,
     unix_glNormal3hNV,
     unix_glNormal3hvNV,
-    unix_glNormal3x,
     unix_glNormal3xOES,
     unix_glNormal3xvOES,
     unix_glNormalFormatNV,
@@ -27734,9 +27181,7 @@
     unix_glObjectPtrLabel,
     unix_glObjectPurgeableAPPLE,
     unix_glObjectUnpurgeableAPPLE,
-    unix_glOrthof,
     unix_glOrthofOES,
-    unix_glOrthox,
     unix_glOrthoxOES,
     unix_glPNTrianglesfATI,
     unix_glPNTrianglesiATI,
@@ -27794,17 +27239,13 @@
     unix_glPointParameteriNV,
     unix_glPointParameteriv,
     unix_glPointParameterivNV,
-    unix_glPointParameterx,
-    unix_glPointParameterxv,
     unix_glPointParameterxvOES,
-    unix_glPointSizex,
     unix_glPointSizexOES,
     unix_glPollAsyncSGIX,
     unix_glPollInstrumentsSGIX,
     unix_glPolygonOffsetClamp,
     unix_glPolygonOffsetClampEXT,
     unix_glPolygonOffsetEXT,
-    unix_glPolygonOffsetx,
     unix_glPolygonOffsetxOES,
     unix_glPopDebugGroup,
     unix_glPopGroupMarkerEXT,
@@ -28063,11 +27504,9 @@
     unix_glResolveDepthValuesNV,
     unix_glResumeTransformFeedback,
     unix_glResumeTransformFeedbackNV,
-    unix_glRotatex,
     unix_glRotatexOES,
     unix_glSampleCoverage,
     unix_glSampleCoverageARB,
-    unix_glSampleCoveragex,
     unix_glSampleMapATI,
     unix_glSampleMaskEXT,
     unix_glSampleMaskIndexedNV,
@@ -28081,7 +27520,6 @@
     unix_glSamplerParameterfv,
     unix_glSamplerParameteri,
     unix_glSamplerParameteriv,
-    unix_glScalex,
     unix_glScalexOES,
     unix_glScissorArrayv,
     unix_glScissorExclusiveArrayvNV,
@@ -28131,7 +27569,6 @@
     unix_glSelectPerfMonitorCountersAMD,
     unix_glSelectTextureCoordSetSGIS,
     unix_glSelectTextureSGIS,
-    unix_glSemaphoreParameterivNV,
     unix_glSemaphoreParameterui64vEXT,
     unix_glSeparableFilter2D,
     unix_glSeparableFilter2DEXT,
@@ -28260,9 +27697,7 @@
     unix_glTexCoordPointerEXT,
     unix_glTexCoordPointerListIBM,
     unix_glTexCoordPointervINTEL,
-    unix_glTexEnvx,
     unix_glTexEnvxOES,
-    unix_glTexEnvxv,
     unix_glTexEnvxvOES,
     unix_glTexFilterFuncSGIS,
     unix_glTexGenxOES,
@@ -28275,23 +27710,17 @@
     unix_glTexImage3DMultisampleCoverageNV,
     unix_glTexImage4DSGIS,
     unix_glTexPageCommitmentARB,
-    unix_glTexPageCommitmentMemNV,
     unix_glTexParameterIiv,
     unix_glTexParameterIivEXT,
     unix_glTexParameterIuiv,
     unix_glTexParameterIuivEXT,
-    unix_glTexParameterx,
     unix_glTexParameterxOES,
-    unix_glTexParameterxv,
     unix_glTexParameterxvOES,
     unix_glTexRenderbufferNV,
     unix_glTexStorage1D,
-    unix_glTexStorage1DEXT,
     unix_glTexStorage2D,
-    unix_glTexStorage2DEXT,
     unix_glTexStorage2DMultisample,
     unix_glTexStorage3D,
-    unix_glTexStorage3DEXT,
     unix_glTexStorage3DMultisample,
     unix_glTexStorageMem1DEXT,
     unix_glTexStorageMem2DEXT,
@@ -28323,7 +27752,6 @@
     unix_glTextureMaterialEXT,
     unix_glTextureNormalEXT,
     unix_glTexturePageCommitmentEXT,
-    unix_glTexturePageCommitmentMemNV,
     unix_glTextureParameterIiv,
     unix_glTextureParameterIivEXT,
     unix_glTextureParameterIuiv,
@@ -28370,7 +27798,6 @@
     unix_glTransformFeedbackVaryingsEXT,
     unix_glTransformFeedbackVaryingsNV,
     unix_glTransformPathNV,
-    unix_glTranslatex,
     unix_glTranslatexOES,
     unix_glUniform1d,
     unix_glUniform1dv,
@@ -28937,6 +28364,7 @@
     unix_wglCreatePbufferARB,
     unix_wglDestroyPbufferARB,
     unix_wglFreeMemoryNV,
+    unix_wglGetCurrentReadDCARB,
     unix_wglGetExtensionsStringARB,
     unix_wglGetExtensionsStringEXT,
     unix_wglGetPbufferDCARB,
