diff --git a/dlls/opengl32/thunks.c b/dlls/opengl32/thunks.c
--- a/dlls/opengl32/thunks.c
+++ b/dlls/opengl32/thunks.c
@@ -2846,14 +2846,6 @@
     if ((status = UNIX_CALL( glAlphaFragmentOp3ATI, &args ))) WARN( "glAlphaFragmentOp3ATI returned %#lx\n", status );
 }
 
-static void WINAPI glAlphaFuncx( GLenum func, GLfixed ref )
-{
-    struct glAlphaFuncx_params args = { .teb = NtCurrentTeb(), .func = func, .ref = ref };
-    NTSTATUS status;
-    TRACE( "func %d, ref %d\n", func, ref );
-    if ((status = UNIX_CALL( glAlphaFuncx, &args ))) WARN( "glAlphaFuncx returned %#lx\n", status );
-}
-
 static void WINAPI glAlphaFuncxOES( GLenum func, GLfixed ref )
 {
     struct glAlphaFuncxOES_params args = { .teb = NtCurrentTeb(), .func = func, .ref = ref };
@@ -3791,22 +3783,6 @@
     if ((status = UNIX_CALL( glBlitFramebufferEXT, &args ))) WARN( "glBlitFramebufferEXT returned %#lx\n", status );
 }
 
-static void WINAPI glBlitFramebufferLayerEXT( GLint srcX0, GLint srcY0, GLint srcX1, GLint srcY1, GLint srcLayer, GLint dstX0, GLint dstY0, GLint dstX1, GLint dstY1, GLint dstLayer, GLbitfield mask, GLenum filter )
-{
-    struct glBlitFramebufferLayerEXT_params args = { .teb = NtCurrentTeb(), .srcX0 = srcX0, .srcY0 = srcY0, .srcX1 = srcX1, .srcY1 = srcY1, .srcLayer = srcLayer, .dstX0 = dstX0, .dstY0 = dstY0, .dstX1 = dstX1, .dstY1 = dstY1, .dstLayer = dstLayer, .mask = mask, .filter = filter };
-    NTSTATUS status;
-    TRACE( "srcX0 %d, srcY0 %d, srcX1 %d, srcY1 %d, srcLayer %d, dstX0 %d, dstY0 %d, dstX1 %d, dstY1 %d, dstLayer %d, mask %d, filter %d\n", srcX0, srcY0, srcX1, srcY1, srcLayer, dstX0, dstY0, dstX1, dstY1, dstLayer, mask, filter );
-    if ((status = UNIX_CALL( glBlitFramebufferLayerEXT, &args ))) WARN( "glBlitFramebufferLayerEXT returned %#lx\n", status );
-}
-
-static void WINAPI glBlitFramebufferLayersEXT( GLint srcX0, GLint srcY0, GLint srcX1, GLint srcY1, GLint dstX0, GLint dstY0, GLint dstX1, GLint dstY1, GLbitfield mask, GLenum filter )
-{
-    struct glBlitFramebufferLayersEXT_params args = { .teb = NtCurrentTeb(), .srcX0 = srcX0, .srcY0 = srcY0, .srcX1 = srcX1, .srcY1 = srcY1, .dstX0 = dstX0, .dstY0 = dstY0, .dstX1 = dstX1, .dstY1 = dstY1, .mask = mask, .filter = filter };
-    NTSTATUS status;
-    TRACE( "srcX0 %d, srcY0 %d, srcX1 %d, srcY1 %d, dstX0 %d, dstY0 %d, dstX1 %d, dstY1 %d, mask %d, filter %d\n", srcX0, srcY0, srcX1, srcY1, dstX0, dstY0, dstX1, dstY1, mask, filter );
-    if ((status = UNIX_CALL( glBlitFramebufferLayersEXT, &args ))) WARN( "glBlitFramebufferLayersEXT returned %#lx\n", status );
-}
-
 static void WINAPI glBlitNamedFramebuffer( GLuint readFramebuffer, GLuint drawFramebuffer, GLint srcX0, GLint srcY0, GLint srcX1, GLint srcY1, GLint dstX0, GLint dstY0, GLint dstX1, GLint dstY1, GLbitfield mask, GLenum filter )
 {
     struct glBlitNamedFramebuffer_params args = { .teb = NtCurrentTeb(), .readFramebuffer = readFramebuffer, .drawFramebuffer = drawFramebuffer, .srcX0 = srcX0, .srcY0 = srcY0, .srcX1 = srcX1, .srcY1 = srcY1, .dstX0 = dstX0, .dstY0 = dstY0, .dstX1 = dstX1, .dstY1 = dstY1, .mask = mask, .filter = filter };
@@ -3853,14 +3829,6 @@
     NTSTATUS status;
     TRACE( "target %d, offset %Id, size %Id, commit %d\n", target, offset, size, commit );
     if ((status = UNIX_CALL( glBufferPageCommitmentARB, &args ))) WARN( "glBufferPageCommitmentARB returned %#lx\n", status );
-}
-
-static void WINAPI glBufferPageCommitmentMemNV( GLenum target, GLintptr offset, GLsizeiptr size, GLuint memory, GLuint64 memOffset, GLboolean commit )
-{
-    struct glBufferPageCommitmentMemNV_params args = { .teb = NtCurrentTeb(), .target = target, .offset = offset, .size = size, .memory = memory, .memOffset = memOffset, .commit = commit };
-    NTSTATUS status;
-    TRACE( "target %d, offset %Id, size %Id, memory %d, memOffset %s, commit %d\n", target, offset, size, memory, wine_dbgstr_longlong(memOffset), commit );
-    if ((status = UNIX_CALL( glBufferPageCommitmentMemNV, &args ))) WARN( "glBufferPageCommitmentMemNV returned %#lx\n", status );
 }
 
 static void WINAPI glBufferParameteriAPPLE( GLenum target, GLenum pname, GLint param )
@@ -4052,14 +4020,6 @@
     if ((status = UNIX_CALL( glClearColorIuiEXT, &args ))) WARN( "glClearColorIuiEXT returned %#lx\n", status );
 }
 
-static void WINAPI glClearColorx( GLfixed red, GLfixed green, GLfixed blue, GLfixed alpha )
-{
-    struct glClearColorx_params args = { .teb = NtCurrentTeb(), .red = red, .green = green, .blue = blue, .alpha = alpha };
-    NTSTATUS status;
-    TRACE( "red %d, green %d, blue %d, alpha %d\n", red, green, blue, alpha );
-    if ((status = UNIX_CALL( glClearColorx, &args ))) WARN( "glClearColorx returned %#lx\n", status );
-}
-
 static void WINAPI glClearColorxOES( GLfixed red, GLfixed green, GLfixed blue, GLfixed alpha )
 {
     struct glClearColorxOES_params args = { .teb = NtCurrentTeb(), .red = red, .green = green, .blue = blue, .alpha = alpha };
@@ -4090,14 +4050,6 @@
     NTSTATUS status;
     TRACE( "depth %f\n", depth );
     if ((status = UNIX_CALL( glClearDepthfOES, &args ))) WARN( "glClearDepthfOES returned %#lx\n", status );
-}
-
-static void WINAPI glClearDepthx( GLfixed depth )
-{
-    struct glClearDepthx_params args = { .teb = NtCurrentTeb(), .depth = depth };
-    NTSTATUS status;
-    TRACE( "depth %d\n", depth );
-    if ((status = UNIX_CALL( glClearDepthx, &args ))) WARN( "glClearDepthx returned %#lx\n", status );
 }
 
 static void WINAPI glClearDepthxOES( GLfixed depth )
@@ -4245,14 +4197,6 @@
     if ((status = UNIX_CALL( glClipControl, &args ))) WARN( "glClipControl returned %#lx\n", status );
 }
 
-static void WINAPI glClipPlanef( GLenum p, const GLfloat *eqn )
-{
-    struct glClipPlanef_params args = { .teb = NtCurrentTeb(), .p = p, .eqn = eqn };
-    NTSTATUS status;
-    TRACE( "p %d, eqn %p\n", p, eqn );
-    if ((status = UNIX_CALL( glClipPlanef, &args ))) WARN( "glClipPlanef returned %#lx\n", status );
-}
-
 static void WINAPI glClipPlanefOES( GLenum plane, const GLfloat *equation )
 {
     struct glClipPlanefOES_params args = { .teb = NtCurrentTeb(), .plane = plane, .equation = equation };
@@ -4261,14 +4205,6 @@
     if ((status = UNIX_CALL( glClipPlanefOES, &args ))) WARN( "glClipPlanefOES returned %#lx\n", status );
 }
 
-static void WINAPI glClipPlanex( GLenum plane, const GLfixed *equation )
-{
-    struct glClipPlanex_params args = { .teb = NtCurrentTeb(), .plane = plane, .equation = equation };
-    NTSTATUS status;
-    TRACE( "plane %d, equation %p\n", plane, equation );
-    if ((status = UNIX_CALL( glClipPlanex, &args ))) WARN( "glClipPlanex returned %#lx\n", status );
-}
-
 static void WINAPI glClipPlanexOES( GLenum plane, const GLfixed *equation )
 {
     struct glClipPlanexOES_params args = { .teb = NtCurrentTeb(), .plane = plane, .equation = equation };
@@ -4387,14 +4323,6 @@
     NTSTATUS status;
     TRACE( "c %p, v %p\n", c, v );
     if ((status = UNIX_CALL( glColor4ubVertex3fvSUN, &args ))) WARN( "glColor4ubVertex3fvSUN returned %#lx\n", status );
-}
-
-static void WINAPI glColor4x( GLfixed red, GLfixed green, GLfixed blue, GLfixed alpha )
-{
-    struct glColor4x_params args = { .teb = NtCurrentTeb(), .red = red, .green = green, .blue = blue, .alpha = alpha };
-    NTSTATUS status;
-    TRACE( "red %d, green %d, blue %d, alpha %d\n", red, green, blue, alpha );
-    if ((status = UNIX_CALL( glColor4x, &args ))) WARN( "glColor4x returned %#lx\n", status );
 }
 
 static void WINAPI glColor4xOES( GLfixed red, GLfixed green, GLfixed blue, GLfixed alpha )
@@ -5432,14 +5360,6 @@
     if ((status = UNIX_CALL( glCreateSamplers, &args ))) WARN( "glCreateSamplers returned %#lx\n", status );
 }
 
-static void WINAPI glCreateSemaphoresNV( GLsizei n, GLuint *semaphores )
-{
-    struct glCreateSemaphoresNV_params args = { .teb = NtCurrentTeb(), .n = n, .semaphores = semaphores };
-    NTSTATUS status;
-    TRACE( "n %d, semaphores %p\n", n, semaphores );
-    if ((status = UNIX_CALL( glCreateSemaphoresNV, &args ))) WARN( "glCreateSemaphoresNV returned %#lx\n", status );
-}
-
 static GLuint WINAPI glCreateShader( GLenum type )
 {
     struct glCreateShader_params args = { .teb = NtCurrentTeb(), .type = type };
@@ -6021,14 +5941,6 @@
     if ((status = UNIX_CALL( glDepthRangefOES, &args ))) WARN( "glDepthRangefOES returned %#lx\n", status );
 }
 
-static void WINAPI glDepthRangex( GLfixed n, GLfixed f )
-{
-    struct glDepthRangex_params args = { .teb = NtCurrentTeb(), .n = n, .f = f };
-    NTSTATUS status;
-    TRACE( "n %d, f %d\n", n, f );
-    if ((status = UNIX_CALL( glDepthRangex, &args ))) WARN( "glDepthRangex returned %#lx\n", status );
-}
-
 static void WINAPI glDepthRangexOES( GLfixed n, GLfixed f )
 {
     struct glDepthRangexOES_params args = { .teb = NtCurrentTeb(), .n = n, .f = f };
@@ -6371,22 +6283,6 @@
     NTSTATUS status;
     TRACE( "mode %d, first %d, count %d, width %d\n", mode, first, count, width );
     if ((status = UNIX_CALL( glDrawMeshArraysSUN, &args ))) WARN( "glDrawMeshArraysSUN returned %#lx\n", status );
-}
-
-static void WINAPI glDrawMeshTasksEXT( GLuint num_groups_x, GLuint num_groups_y, GLuint num_groups_z )
-{
-    struct glDrawMeshTasksEXT_params args = { .teb = NtCurrentTeb(), .num_groups_x = num_groups_x, .num_groups_y = num_groups_y, .num_groups_z = num_groups_z };
-    NTSTATUS status;
-    TRACE( "num_groups_x %d, num_groups_y %d, num_groups_z %d\n", num_groups_x, num_groups_y, num_groups_z );
-    if ((status = UNIX_CALL( glDrawMeshTasksEXT, &args ))) WARN( "glDrawMeshTasksEXT returned %#lx\n", status );
-}
-
-static void WINAPI glDrawMeshTasksIndirectEXT( GLintptr indirect )
-{
-    struct glDrawMeshTasksIndirectEXT_params args = { .teb = NtCurrentTeb(), .indirect = indirect };
-    NTSTATUS status;
-    TRACE( "indirect %Id\n", indirect );
-    if ((status = UNIX_CALL( glDrawMeshTasksIndirectEXT, &args ))) WARN( "glDrawMeshTasksIndirectEXT returned %#lx\n", status );
 }
 
 static void WINAPI glDrawMeshTasksIndirectNV( GLintptr indirect )
@@ -7087,14 +6983,6 @@
     if ((status = UNIX_CALL( glFogFuncSGIS, &args ))) WARN( "glFogFuncSGIS returned %#lx\n", status );
 }
 
-static void WINAPI glFogx( GLenum pname, GLfixed param )
-{
-    struct glFogx_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "pname %d, param %d\n", pname, param );
-    if ((status = UNIX_CALL( glFogx, &args ))) WARN( "glFogx returned %#lx\n", status );
-}
-
 static void WINAPI glFogxOES( GLenum pname, GLfixed param )
 {
     struct glFogxOES_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
@@ -7103,14 +6991,6 @@
     if ((status = UNIX_CALL( glFogxOES, &args ))) WARN( "glFogxOES returned %#lx\n", status );
 }
 
-static void WINAPI glFogxv( GLenum pname, const GLfixed *param )
-{
-    struct glFogxv_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "pname %d, param %p\n", pname, param );
-    if ((status = UNIX_CALL( glFogxv, &args ))) WARN( "glFogxv returned %#lx\n", status );
-}
-
 static void WINAPI glFogxvOES( GLenum pname, const GLfixed *param )
 {
     struct glFogxvOES_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
@@ -7463,28 +7343,12 @@
     if ((status = UNIX_CALL( glFreeObjectBufferATI, &args ))) WARN( "glFreeObjectBufferATI returned %#lx\n", status );
 }
 
-static void WINAPI glFrustumf( GLfloat l, GLfloat r, GLfloat b, GLfloat t, GLfloat n, GLfloat f )
-{
-    struct glFrustumf_params args = { .teb = NtCurrentTeb(), .l = l, .r = r, .b = b, .t = t, .n = n, .f = f };
-    NTSTATUS status;
-    TRACE( "l %f, r %f, b %f, t %f, n %f, f %f\n", l, r, b, t, n, f );
-    if ((status = UNIX_CALL( glFrustumf, &args ))) WARN( "glFrustumf returned %#lx\n", status );
-}
-
 static void WINAPI glFrustumfOES( GLfloat l, GLfloat r, GLfloat b, GLfloat t, GLfloat n, GLfloat f )
 {
     struct glFrustumfOES_params args = { .teb = NtCurrentTeb(), .l = l, .r = r, .b = b, .t = t, .n = n, .f = f };
     NTSTATUS status;
     TRACE( "l %f, r %f, b %f, t %f, n %f, f %f\n", l, r, b, t, n, f );
     if ((status = UNIX_CALL( glFrustumfOES, &args ))) WARN( "glFrustumfOES returned %#lx\n", status );
-}
-
-static void WINAPI glFrustumx( GLfixed l, GLfixed r, GLfixed b, GLfixed t, GLfixed n, GLfixed f )
-{
-    struct glFrustumx_params args = { .teb = NtCurrentTeb(), .l = l, .r = r, .b = b, .t = t, .n = n, .f = f };
-    NTSTATUS status;
-    TRACE( "l %d, r %d, b %d, t %d, n %d, f %d\n", l, r, b, t, n, f );
-    if ((status = UNIX_CALL( glFrustumx, &args ))) WARN( "glFrustumx returned %#lx\n", status );
 }
 
 static void WINAPI glFrustumxOES( GLfixed l, GLfixed r, GLfixed b, GLfixed t, GLfixed n, GLfixed f )
@@ -8006,28 +7870,12 @@
     if ((status = UNIX_CALL( glGetBufferSubDataARB, &args ))) WARN( "glGetBufferSubDataARB returned %#lx\n", status );
 }
 
-static void WINAPI glGetClipPlanef( GLenum plane, GLfloat *equation )
-{
-    struct glGetClipPlanef_params args = { .teb = NtCurrentTeb(), .plane = plane, .equation = equation };
-    NTSTATUS status;
-    TRACE( "plane %d, equation %p\n", plane, equation );
-    if ((status = UNIX_CALL( glGetClipPlanef, &args ))) WARN( "glGetClipPlanef returned %#lx\n", status );
-}
-
 static void WINAPI glGetClipPlanefOES( GLenum plane, GLfloat *equation )
 {
     struct glGetClipPlanefOES_params args = { .teb = NtCurrentTeb(), .plane = plane, .equation = equation };
     NTSTATUS status;
     TRACE( "plane %d, equation %p\n", plane, equation );
     if ((status = UNIX_CALL( glGetClipPlanefOES, &args ))) WARN( "glGetClipPlanefOES returned %#lx\n", status );
-}
-
-static void WINAPI glGetClipPlanex( GLenum plane, GLfixed *equation )
-{
-    struct glGetClipPlanex_params args = { .teb = NtCurrentTeb(), .plane = plane, .equation = equation };
-    NTSTATUS status;
-    TRACE( "plane %d, equation %p\n", plane, equation );
-    if ((status = UNIX_CALL( glGetClipPlanex, &args ))) WARN( "glGetClipPlanex returned %#lx\n", status );
 }
 
 static void WINAPI glGetClipPlanexOES( GLenum plane, GLfixed *equation )
@@ -8280,7 +8128,7 @@
     return args.ret;
 }
 
-static GLuint WINAPI glGetDebugMessageLogAMD( GLuint count, GLsizei bufSize, GLenum *categories, GLenum *severities, GLuint *ids, GLsizei *lengths, GLchar *message )
+static GLuint WINAPI glGetDebugMessageLogAMD( GLuint count, GLsizei bufSize, GLenum *categories, GLuint *severities, GLuint *ids, GLsizei *lengths, GLchar *message )
 {
     struct glGetDebugMessageLogAMD_params args = { .teb = NtCurrentTeb(), .count = count, .bufSize = bufSize, .categories = categories, .severities = severities, .ids = ids, .lengths = lengths, .message = message };
     NTSTATUS status;
@@ -8362,14 +8210,6 @@
     if ((status = UNIX_CALL( glGetFirstPerfQueryIdINTEL, &args ))) WARN( "glGetFirstPerfQueryIdINTEL returned %#lx\n", status );
 }
 
-static void WINAPI glGetFixedv( GLenum pname, GLfixed *params )
-{
-    struct glGetFixedv_params args = { .teb = NtCurrentTeb(), .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "pname %d, params %p\n", pname, params );
-    if ((status = UNIX_CALL( glGetFixedv, &args ))) WARN( "glGetFixedv returned %#lx\n", status );
-}
-
 static void WINAPI glGetFixedvOES( GLenum pname, GLfixed *params )
 {
     struct glGetFixedvOES_params args = { .teb = NtCurrentTeb(), .pname = pname, .params = params };
@@ -8755,14 +8595,6 @@
     if ((status = UNIX_CALL( glGetLightxOES, &args ))) WARN( "glGetLightxOES returned %#lx\n", status );
 }
 
-static void WINAPI glGetLightxv( GLenum light, GLenum pname, GLfixed *params )
-{
-    struct glGetLightxv_params args = { .teb = NtCurrentTeb(), .light = light, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "light %d, pname %d, params %p\n", light, pname, params );
-    if ((status = UNIX_CALL( glGetLightxv, &args ))) WARN( "glGetLightxv returned %#lx\n", status );
-}
-
 static void WINAPI glGetListParameterfvSGIX( GLuint list, GLenum pname, GLfloat *params )
 {
     struct glGetListParameterfvSGIX_params args = { .teb = NtCurrentTeb(), .list = list, .pname = pname, .params = params };
@@ -8857,14 +8689,6 @@
     NTSTATUS status;
     TRACE( "face %d, pname %d, param %d\n", face, pname, param );
     if ((status = UNIX_CALL( glGetMaterialxOES, &args ))) WARN( "glGetMaterialxOES returned %#lx\n", status );
-}
-
-static void WINAPI glGetMaterialxv( GLenum face, GLenum pname, GLfixed *params )
-{
-    struct glGetMaterialxv_params args = { .teb = NtCurrentTeb(), .face = face, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "face %d, pname %d, params %p\n", face, pname, params );
-    if ((status = UNIX_CALL( glGetMaterialxv, &args ))) WARN( "glGetMaterialxv returned %#lx\n", status );
 }
 
 static void WINAPI glGetMemoryObjectDetachedResourcesuivNV( GLuint memory, GLenum pname, GLint first, GLsizei count, GLuint *params )
@@ -9975,14 +9799,6 @@
     if ((status = UNIX_CALL( glGetSamplerParameteriv, &args ))) WARN( "glGetSamplerParameteriv returned %#lx\n", status );
 }
 
-static void WINAPI glGetSemaphoreParameterivNV( GLuint semaphore, GLenum pname, GLint *params )
-{
-    struct glGetSemaphoreParameterivNV_params args = { .teb = NtCurrentTeb(), .semaphore = semaphore, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "semaphore %d, pname %d, params %p\n", semaphore, pname, params );
-    if ((status = UNIX_CALL( glGetSemaphoreParameterivNV, &args ))) WARN( "glGetSemaphoreParameterivNV returned %#lx\n", status );
-}
-
 static void WINAPI glGetSemaphoreParameterui64vEXT( GLuint semaphore, GLenum pname, GLuint64 *params )
 {
     struct glGetSemaphoreParameterui64vEXT_params args = { .teb = NtCurrentTeb(), .semaphore = semaphore, .pname = pname, .params = params };
@@ -10122,14 +9938,6 @@
     if ((status = UNIX_CALL( glGetTexBumpParameterivATI, &args ))) WARN( "glGetTexBumpParameterivATI returned %#lx\n", status );
 }
 
-static void WINAPI glGetTexEnvxv( GLenum target, GLenum pname, GLfixed *params )
-{
-    struct glGetTexEnvxv_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "target %d, pname %d, params %p\n", target, pname, params );
-    if ((status = UNIX_CALL( glGetTexEnvxv, &args ))) WARN( "glGetTexEnvxv returned %#lx\n", status );
-}
-
 static void WINAPI glGetTexEnvxvOES( GLenum target, GLenum pname, GLfixed *params )
 {
     struct glGetTexEnvxvOES_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .params = params };
@@ -10200,14 +10008,6 @@
     NTSTATUS status;
     TRACE( "target %d, pname %d, params %p\n", target, pname, params );
     if ((status = UNIX_CALL( glGetTexParameterPointervAPPLE, &args ))) WARN( "glGetTexParameterPointervAPPLE returned %#lx\n", status );
-}
-
-static void WINAPI glGetTexParameterxv( GLenum target, GLenum pname, GLfixed *params )
-{
-    struct glGetTexParameterxv_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "target %d, pname %d, params %p\n", target, pname, params );
-    if ((status = UNIX_CALL( glGetTexParameterxv, &args ))) WARN( "glGetTexParameterxv returned %#lx\n", status );
 }
 
 static void WINAPI glGetTexParameterxvOES( GLenum target, GLenum pname, GLfixed *params )
@@ -12042,14 +11842,6 @@
     if ((status = UNIX_CALL( glLightEnviSGIX, &args ))) WARN( "glLightEnviSGIX returned %#lx\n", status );
 }
 
-static void WINAPI glLightModelx( GLenum pname, GLfixed param )
-{
-    struct glLightModelx_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "pname %d, param %d\n", pname, param );
-    if ((status = UNIX_CALL( glLightModelx, &args ))) WARN( "glLightModelx returned %#lx\n", status );
-}
-
 static void WINAPI glLightModelxOES( GLenum pname, GLfixed param )
 {
     struct glLightModelxOES_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
@@ -12058,14 +11850,6 @@
     if ((status = UNIX_CALL( glLightModelxOES, &args ))) WARN( "glLightModelxOES returned %#lx\n", status );
 }
 
-static void WINAPI glLightModelxv( GLenum pname, const GLfixed *param )
-{
-    struct glLightModelxv_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "pname %d, param %p\n", pname, param );
-    if ((status = UNIX_CALL( glLightModelxv, &args ))) WARN( "glLightModelxv returned %#lx\n", status );
-}
-
 static void WINAPI glLightModelxvOES( GLenum pname, const GLfixed *param )
 {
     struct glLightModelxvOES_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
@@ -12074,14 +11858,6 @@
     if ((status = UNIX_CALL( glLightModelxvOES, &args ))) WARN( "glLightModelxvOES returned %#lx\n", status );
 }
 
-static void WINAPI glLightx( GLenum light, GLenum pname, GLfixed param )
-{
-    struct glLightx_params args = { .teb = NtCurrentTeb(), .light = light, .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "light %d, pname %d, param %d\n", light, pname, param );
-    if ((status = UNIX_CALL( glLightx, &args ))) WARN( "glLightx returned %#lx\n", status );
-}
-
 static void WINAPI glLightxOES( GLenum light, GLenum pname, GLfixed param )
 {
     struct glLightxOES_params args = { .teb = NtCurrentTeb(), .light = light, .pname = pname, .param = param };
@@ -12090,14 +11866,6 @@
     if ((status = UNIX_CALL( glLightxOES, &args ))) WARN( "glLightxOES returned %#lx\n", status );
 }
 
-static void WINAPI glLightxv( GLenum light, GLenum pname, const GLfixed *params )
-{
-    struct glLightxv_params args = { .teb = NtCurrentTeb(), .light = light, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "light %d, pname %d, params %p\n", light, pname, params );
-    if ((status = UNIX_CALL( glLightxv, &args ))) WARN( "glLightxv returned %#lx\n", status );
-}
-
 static void WINAPI glLightxvOES( GLenum light, GLenum pname, const GLfixed *params )
 {
     struct glLightxvOES_params args = { .teb = NtCurrentTeb(), .light = light, .pname = pname, .params = params };
@@ -12106,14 +11874,6 @@
     if ((status = UNIX_CALL( glLightxvOES, &args ))) WARN( "glLightxvOES returned %#lx\n", status );
 }
 
-static void WINAPI glLineWidthx( GLfixed width )
-{
-    struct glLineWidthx_params args = { .teb = NtCurrentTeb(), .width = width };
-    NTSTATUS status;
-    TRACE( "width %d\n", width );
-    if ((status = UNIX_CALL( glLineWidthx, &args ))) WARN( "glLineWidthx returned %#lx\n", status );
-}
-
 static void WINAPI glLineWidthxOES( GLfixed width )
 {
     struct glLineWidthxOES_params args = { .teb = NtCurrentTeb(), .width = width };
@@ -12186,14 +11946,6 @@
     if ((status = UNIX_CALL( glLoadIdentityDeformationMapSGIX, &args ))) WARN( "glLoadIdentityDeformationMapSGIX returned %#lx\n", status );
 }
 
-static void WINAPI glLoadMatrixx( const GLfixed *m )
-{
-    struct glLoadMatrixx_params args = { .teb = NtCurrentTeb(), .m = m };
-    NTSTATUS status;
-    TRACE( "m %p\n", m );
-    if ((status = UNIX_CALL( glLoadMatrixx, &args ))) WARN( "glLoadMatrixx returned %#lx\n", status );
-}
-
 static void WINAPI glLoadMatrixxOES( const GLfixed *m )
 {
     struct glLoadMatrixxOES_params args = { .teb = NtCurrentTeb(), .m = m };
@@ -12386,33 +12138,6 @@
     if ((status = UNIX_CALL( glMap2xOES, &args ))) WARN( "glMap2xOES returned %#lx\n", status );
 }
 
-static void * WINAPI glMapBuffer( GLenum target, GLenum access )
-{
-    struct glMapBuffer_params args = { .teb = NtCurrentTeb(), .target = target, .access = access };
-    NTSTATUS status;
-    TRACE( "target %d, access %d\n", target, access );
-    if ((status = UNIX_CALL( glMapBuffer, &args ))) WARN( "glMapBuffer returned %#lx\n", status );
-    return args.ret;
-}
-
-static void * WINAPI glMapBufferARB( GLenum target, GLenum access )
-{
-    struct glMapBufferARB_params args = { .teb = NtCurrentTeb(), .target = target, .access = access };
-    NTSTATUS status;
-    TRACE( "target %d, access %d\n", target, access );
-    if ((status = UNIX_CALL( glMapBufferARB, &args ))) WARN( "glMapBufferARB returned %#lx\n", status );
-    return args.ret;
-}
-
-static void * WINAPI glMapBufferRange( GLenum target, GLintptr offset, GLsizeiptr length, GLbitfield access )
-{
-    struct glMapBufferRange_params args = { .teb = NtCurrentTeb(), .target = target, .offset = offset, .length = length, .access = access };
-    NTSTATUS status;
-    TRACE( "target %d, offset %Id, length %Id, access %d\n", target, offset, length, access );
-    if ((status = UNIX_CALL( glMapBufferRange, &args ))) WARN( "glMapBufferRange returned %#lx\n", status );
-    return args.ret;
-}
-
 static void WINAPI glMapControlPointsNV( GLenum target, GLuint index, GLenum type, GLsizei ustride, GLsizei vstride, GLint uorder, GLint vorder, GLboolean packed, const void *points )
 {
     struct glMapControlPointsNV_params args = { .teb = NtCurrentTeb(), .target = target, .index = index, .type = type, .ustride = ustride, .vstride = vstride, .uorder = uorder, .vorder = vorder, .packed = packed, .points = points };
@@ -12435,42 +12160,6 @@
     NTSTATUS status;
     TRACE( "n %d, u1 %d, u2 %d, v1 %d, v2 %d\n", n, u1, u2, v1, v2 );
     if ((status = UNIX_CALL( glMapGrid2xOES, &args ))) WARN( "glMapGrid2xOES returned %#lx\n", status );
-}
-
-static void * WINAPI glMapNamedBuffer( GLuint buffer, GLenum access )
-{
-    struct glMapNamedBuffer_params args = { .teb = NtCurrentTeb(), .buffer = buffer, .access = access };
-    NTSTATUS status;
-    TRACE( "buffer %d, access %d\n", buffer, access );
-    if ((status = UNIX_CALL( glMapNamedBuffer, &args ))) WARN( "glMapNamedBuffer returned %#lx\n", status );
-    return args.ret;
-}
-
-static void * WINAPI glMapNamedBufferEXT( GLuint buffer, GLenum access )
-{
-    struct glMapNamedBufferEXT_params args = { .teb = NtCurrentTeb(), .buffer = buffer, .access = access };
-    NTSTATUS status;
-    TRACE( "buffer %d, access %d\n", buffer, access );
-    if ((status = UNIX_CALL( glMapNamedBufferEXT, &args ))) WARN( "glMapNamedBufferEXT returned %#lx\n", status );
-    return args.ret;
-}
-
-static void * WINAPI glMapNamedBufferRange( GLuint buffer, GLintptr offset, GLsizeiptr length, GLbitfield access )
-{
-    struct glMapNamedBufferRange_params args = { .teb = NtCurrentTeb(), .buffer = buffer, .offset = offset, .length = length, .access = access };
-    NTSTATUS status;
-    TRACE( "buffer %d, offset %Id, length %Id, access %d\n", buffer, offset, length, access );
-    if ((status = UNIX_CALL( glMapNamedBufferRange, &args ))) WARN( "glMapNamedBufferRange returned %#lx\n", status );
-    return args.ret;
-}
-
-static void * WINAPI glMapNamedBufferRangeEXT( GLuint buffer, GLintptr offset, GLsizeiptr length, GLbitfield access )
-{
-    struct glMapNamedBufferRangeEXT_params args = { .teb = NtCurrentTeb(), .buffer = buffer, .offset = offset, .length = length, .access = access };
-    NTSTATUS status;
-    TRACE( "buffer %d, offset %Id, length %Id, access %d\n", buffer, offset, length, access );
-    if ((status = UNIX_CALL( glMapNamedBufferRangeEXT, &args ))) WARN( "glMapNamedBufferRangeEXT returned %#lx\n", status );
-    return args.ret;
 }
 
 static void * WINAPI glMapObjectBufferATI( GLuint buffer )
@@ -12539,14 +12228,6 @@
     if ((status = UNIX_CALL( glMapVertexAttrib2fAPPLE, &args ))) WARN( "glMapVertexAttrib2fAPPLE returned %#lx\n", status );
 }
 
-static void WINAPI glMaterialx( GLenum face, GLenum pname, GLfixed param )
-{
-    struct glMaterialx_params args = { .teb = NtCurrentTeb(), .face = face, .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "face %d, pname %d, param %d\n", face, pname, param );
-    if ((status = UNIX_CALL( glMaterialx, &args ))) WARN( "glMaterialx returned %#lx\n", status );
-}
-
 static void WINAPI glMaterialxOES( GLenum face, GLenum pname, GLfixed param )
 {
     struct glMaterialxOES_params args = { .teb = NtCurrentTeb(), .face = face, .pname = pname, .param = param };
@@ -12555,14 +12236,6 @@
     if ((status = UNIX_CALL( glMaterialxOES, &args ))) WARN( "glMaterialxOES returned %#lx\n", status );
 }
 
-static void WINAPI glMaterialxv( GLenum face, GLenum pname, const GLfixed *param )
-{
-    struct glMaterialxv_params args = { .teb = NtCurrentTeb(), .face = face, .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "face %d, pname %d, param %p\n", face, pname, param );
-    if ((status = UNIX_CALL( glMaterialxv, &args ))) WARN( "glMaterialxv returned %#lx\n", status );
-}
-
 static void WINAPI glMaterialxvOES( GLenum face, GLenum pname, const GLfixed *param )
 {
     struct glMaterialxvOES_params args = { .teb = NtCurrentTeb(), .face = face, .pname = pname, .param = param };
@@ -12883,14 +12556,6 @@
     if ((status = UNIX_CALL( glMinmaxEXT, &args ))) WARN( "glMinmaxEXT returned %#lx\n", status );
 }
 
-static void WINAPI glMultMatrixx( const GLfixed *m )
-{
-    struct glMultMatrixx_params args = { .teb = NtCurrentTeb(), .m = m };
-    NTSTATUS status;
-    TRACE( "m %p\n", m );
-    if ((status = UNIX_CALL( glMultMatrixx, &args ))) WARN( "glMultMatrixx returned %#lx\n", status );
-}
-
 static void WINAPI glMultMatrixxOES( const GLfixed *m )
 {
     struct glMultMatrixxOES_params args = { .teb = NtCurrentTeb(), .m = m };
@@ -13083,14 +12748,6 @@
     if ((status = UNIX_CALL( glMultiDrawElementsIndirectCountARB, &args ))) WARN( "glMultiDrawElementsIndirectCountARB returned %#lx\n", status );
 }
 
-static void WINAPI glMultiDrawMeshTasksIndirectCountEXT( GLintptr indirect, GLintptr drawcount, GLsizei maxdrawcount, GLsizei stride )
-{
-    struct glMultiDrawMeshTasksIndirectCountEXT_params args = { .teb = NtCurrentTeb(), .indirect = indirect, .drawcount = drawcount, .maxdrawcount = maxdrawcount, .stride = stride };
-    NTSTATUS status;
-    TRACE( "indirect %Id, drawcount %Id, maxdrawcount %d, stride %d\n", indirect, drawcount, maxdrawcount, stride );
-    if ((status = UNIX_CALL( glMultiDrawMeshTasksIndirectCountEXT, &args ))) WARN( "glMultiDrawMeshTasksIndirectCountEXT returned %#lx\n", status );
-}
-
 static void WINAPI glMultiDrawMeshTasksIndirectCountNV( GLintptr indirect, GLintptr drawcount, GLsizei maxdrawcount, GLsizei stride )
 {
     struct glMultiDrawMeshTasksIndirectCountNV_params args = { .teb = NtCurrentTeb(), .indirect = indirect, .drawcount = drawcount, .maxdrawcount = maxdrawcount, .stride = stride };
@@ -13099,14 +12756,6 @@
     if ((status = UNIX_CALL( glMultiDrawMeshTasksIndirectCountNV, &args ))) WARN( "glMultiDrawMeshTasksIndirectCountNV returned %#lx\n", status );
 }
 
-static void WINAPI glMultiDrawMeshTasksIndirectEXT( GLintptr indirect, GLsizei drawcount, GLsizei stride )
-{
-    struct glMultiDrawMeshTasksIndirectEXT_params args = { .teb = NtCurrentTeb(), .indirect = indirect, .drawcount = drawcount, .stride = stride };
-    NTSTATUS status;
-    TRACE( "indirect %Id, drawcount %d, stride %d\n", indirect, drawcount, stride );
-    if ((status = UNIX_CALL( glMultiDrawMeshTasksIndirectEXT, &args ))) WARN( "glMultiDrawMeshTasksIndirectEXT returned %#lx\n", status );
-}
-
 static void WINAPI glMultiDrawMeshTasksIndirectNV( GLintptr indirect, GLsizei drawcount, GLsizei stride )
 {
     struct glMultiDrawMeshTasksIndirectNV_params args = { .teb = NtCurrentTeb(), .indirect = indirect, .drawcount = drawcount, .stride = stride };
@@ -14091,14 +13740,6 @@
     if ((status = UNIX_CALL( glMultiTexCoord4svSGIS, &args ))) WARN( "glMultiTexCoord4svSGIS returned %#lx\n", status );
 }
 
-static void WINAPI glMultiTexCoord4x( GLenum texture, GLfixed s, GLfixed t, GLfixed r, GLfixed q )
-{
-    struct glMultiTexCoord4x_params args = { .teb = NtCurrentTeb(), .texture = texture, .s = s, .t = t, .r = r, .q = q };
-    NTSTATUS status;
-    TRACE( "texture %d, s %d, t %d, r %d, q %d\n", texture, s, t, r, q );
-    if ((status = UNIX_CALL( glMultiTexCoord4x, &args ))) WARN( "glMultiTexCoord4x returned %#lx\n", status );
-}
-
 static void WINAPI glMultiTexCoord4xOES( GLenum texture, GLfixed s, GLfixed t, GLfixed r, GLfixed q )
 {
     struct glMultiTexCoord4xOES_params args = { .teb = NtCurrentTeb(), .texture = texture, .s = s, .t = t, .r = r, .q = q };
@@ -14531,14 +14172,6 @@
     if ((status = UNIX_CALL( glNamedBufferPageCommitmentEXT, &args ))) WARN( "glNamedBufferPageCommitmentEXT returned %#lx\n", status );
 }
 
-static void WINAPI glNamedBufferPageCommitmentMemNV( GLuint buffer, GLintptr offset, GLsizeiptr size, GLuint memory, GLuint64 memOffset, GLboolean commit )
-{
-    struct glNamedBufferPageCommitmentMemNV_params args = { .teb = NtCurrentTeb(), .buffer = buffer, .offset = offset, .size = size, .memory = memory, .memOffset = memOffset, .commit = commit };
-    NTSTATUS status;
-    TRACE( "buffer %d, offset %Id, size %Id, memory %d, memOffset %s, commit %d\n", buffer, offset, size, memory, wine_dbgstr_longlong(memOffset), commit );
-    if ((status = UNIX_CALL( glNamedBufferPageCommitmentMemNV, &args ))) WARN( "glNamedBufferPageCommitmentMemNV returned %#lx\n", status );
-}
-
 static void WINAPI glNamedBufferStorage( GLuint buffer, GLsizeiptr size, const void *data, GLbitfield flags )
 {
     struct glNamedBufferStorage_params args = { .teb = NtCurrentTeb(), .buffer = buffer, .size = size, .data = data, .flags = flags };
@@ -14737,14 +14370,6 @@
     NTSTATUS status;
     TRACE( "framebuffer %d, attachment %d, texture %d, level %d, layer %d\n", framebuffer, attachment, texture, level, layer );
     if ((status = UNIX_CALL( glNamedFramebufferTextureLayerEXT, &args ))) WARN( "glNamedFramebufferTextureLayerEXT returned %#lx\n", status );
-}
-
-static void WINAPI glNamedFramebufferTextureMultiviewOVR( GLuint framebuffer, GLenum attachment, GLuint texture, GLint level, GLint baseViewIndex, GLsizei numViews )
-{
-    struct glNamedFramebufferTextureMultiviewOVR_params args = { .teb = NtCurrentTeb(), .framebuffer = framebuffer, .attachment = attachment, .texture = texture, .level = level, .baseViewIndex = baseViewIndex, .numViews = numViews };
-    NTSTATUS status;
-    TRACE( "framebuffer %d, attachment %d, texture %d, level %d, baseViewIndex %d, numViews %d\n", framebuffer, attachment, texture, level, baseViewIndex, numViews );
-    if ((status = UNIX_CALL( glNamedFramebufferTextureMultiviewOVR, &args ))) WARN( "glNamedFramebufferTextureMultiviewOVR returned %#lx\n", status );
 }
 
 static void WINAPI glNamedProgramLocalParameter4dEXT( GLuint program, GLenum target, GLuint index, GLdouble x, GLdouble y, GLdouble z, GLdouble w )
@@ -14949,14 +14574,6 @@
     if ((status = UNIX_CALL( glNormal3hvNV, &args ))) WARN( "glNormal3hvNV returned %#lx\n", status );
 }
 
-static void WINAPI glNormal3x( GLfixed nx, GLfixed ny, GLfixed nz )
-{
-    struct glNormal3x_params args = { .teb = NtCurrentTeb(), .nx = nx, .ny = ny, .nz = nz };
-    NTSTATUS status;
-    TRACE( "nx %d, ny %d, nz %d\n", nx, ny, nz );
-    if ((status = UNIX_CALL( glNormal3x, &args ))) WARN( "glNormal3x returned %#lx\n", status );
-}
-
 static void WINAPI glNormal3xOES( GLfixed nx, GLfixed ny, GLfixed nz )
 {
     struct glNormal3xOES_params args = { .teb = NtCurrentTeb(), .nx = nx, .ny = ny, .nz = nz };
@@ -15135,28 +14752,12 @@
     return args.ret;
 }
 
-static void WINAPI glOrthof( GLfloat l, GLfloat r, GLfloat b, GLfloat t, GLfloat n, GLfloat f )
-{
-    struct glOrthof_params args = { .teb = NtCurrentTeb(), .l = l, .r = r, .b = b, .t = t, .n = n, .f = f };
-    NTSTATUS status;
-    TRACE( "l %f, r %f, b %f, t %f, n %f, f %f\n", l, r, b, t, n, f );
-    if ((status = UNIX_CALL( glOrthof, &args ))) WARN( "glOrthof returned %#lx\n", status );
-}
-
 static void WINAPI glOrthofOES( GLfloat l, GLfloat r, GLfloat b, GLfloat t, GLfloat n, GLfloat f )
 {
     struct glOrthofOES_params args = { .teb = NtCurrentTeb(), .l = l, .r = r, .b = b, .t = t, .n = n, .f = f };
     NTSTATUS status;
     TRACE( "l %f, r %f, b %f, t %f, n %f, f %f\n", l, r, b, t, n, f );
     if ((status = UNIX_CALL( glOrthofOES, &args ))) WARN( "glOrthofOES returned %#lx\n", status );
-}
-
-static void WINAPI glOrthox( GLfixed l, GLfixed r, GLfixed b, GLfixed t, GLfixed n, GLfixed f )
-{
-    struct glOrthox_params args = { .teb = NtCurrentTeb(), .l = l, .r = r, .b = b, .t = t, .n = n, .f = f };
-    NTSTATUS status;
-    TRACE( "l %d, r %d, b %d, t %d, n %d, f %d\n", l, r, b, t, n, f );
-    if ((status = UNIX_CALL( glOrthox, &args ))) WARN( "glOrthox returned %#lx\n", status );
 }
 
 static void WINAPI glOrthoxOES( GLfixed l, GLfixed r, GLfixed b, GLfixed t, GLfixed n, GLfixed f )
@@ -15272,10 +14873,11 @@
     return args.ret;
 }
 
-static GLenum WINAPI glPathGlyphIndexRangeNV( GLenum fontTarget, const void *fontName, GLbitfield fontStyle, GLuint pathParameterTemplate, GLfloat emScale, GLuint *baseAndCount )
-{
-    struct glPathGlyphIndexRangeNV_params args = { .teb = NtCurrentTeb(), .fontTarget = fontTarget, .fontName = fontName, .fontStyle = fontStyle, .pathParameterTemplate = pathParameterTemplate, .emScale = emScale, .baseAndCount = baseAndCount };
-    NTSTATUS status;
+static GLenum WINAPI glPathGlyphIndexRangeNV( GLenum fontTarget, const void *fontName, GLbitfield fontStyle, GLuint pathParameterTemplate, GLfloat emScale, GLuint baseAndCount[2] )
+{
+    struct glPathGlyphIndexRangeNV_params args = { .teb = NtCurrentTeb(), .fontTarget = fontTarget, .fontName = fontName, .fontStyle = fontStyle, .pathParameterTemplate = pathParameterTemplate, .emScale = emScale };
+    NTSTATUS status;
+    memcpy( args.baseAndCount, baseAndCount, sizeof(args.baseAndCount) );
     TRACE( "fontTarget %d, fontName %p, fontStyle %d, pathParameterTemplate %d, emScale %f, baseAndCount %p\n", fontTarget, fontName, fontStyle, pathParameterTemplate, emScale, baseAndCount );
     if ((status = UNIX_CALL( glPathGlyphIndexRangeNV, &args ))) WARN( "glPathGlyphIndexRangeNV returned %#lx\n", status );
     return args.ret;
@@ -15619,36 +15221,12 @@
     if ((status = UNIX_CALL( glPointParameterivNV, &args ))) WARN( "glPointParameterivNV returned %#lx\n", status );
 }
 
-static void WINAPI glPointParameterx( GLenum pname, GLfixed param )
-{
-    struct glPointParameterx_params args = { .teb = NtCurrentTeb(), .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "pname %d, param %d\n", pname, param );
-    if ((status = UNIX_CALL( glPointParameterx, &args ))) WARN( "glPointParameterx returned %#lx\n", status );
-}
-
-static void WINAPI glPointParameterxv( GLenum pname, const GLfixed *params )
-{
-    struct glPointParameterxv_params args = { .teb = NtCurrentTeb(), .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "pname %d, params %p\n", pname, params );
-    if ((status = UNIX_CALL( glPointParameterxv, &args ))) WARN( "glPointParameterxv returned %#lx\n", status );
-}
-
 static void WINAPI glPointParameterxvOES( GLenum pname, const GLfixed *params )
 {
     struct glPointParameterxvOES_params args = { .teb = NtCurrentTeb(), .pname = pname, .params = params };
     NTSTATUS status;
     TRACE( "pname %d, params %p\n", pname, params );
     if ((status = UNIX_CALL( glPointParameterxvOES, &args ))) WARN( "glPointParameterxvOES returned %#lx\n", status );
-}
-
-static void WINAPI glPointSizex( GLfixed size )
-{
-    struct glPointSizex_params args = { .teb = NtCurrentTeb(), .size = size };
-    NTSTATUS status;
-    TRACE( "size %d\n", size );
-    if ((status = UNIX_CALL( glPointSizex, &args ))) WARN( "glPointSizex returned %#lx\n", status );
 }
 
 static void WINAPI glPointSizexOES( GLfixed size )
@@ -15701,14 +15279,6 @@
     if ((status = UNIX_CALL( glPolygonOffsetEXT, &args ))) WARN( "glPolygonOffsetEXT returned %#lx\n", status );
 }
 
-static void WINAPI glPolygonOffsetx( GLfixed factor, GLfixed units )
-{
-    struct glPolygonOffsetx_params args = { .teb = NtCurrentTeb(), .factor = factor, .units = units };
-    NTSTATUS status;
-    TRACE( "factor %d, units %d\n", factor, units );
-    if ((status = UNIX_CALL( glPolygonOffsetx, &args ))) WARN( "glPolygonOffsetx returned %#lx\n", status );
-}
-
 static void WINAPI glPolygonOffsetxOES( GLfixed factor, GLfixed units )
 {
     struct glPolygonOffsetxOES_params args = { .teb = NtCurrentTeb(), .factor = factor, .units = units };
@@ -17776,14 +17346,6 @@
     if ((status = UNIX_CALL( glResumeTransformFeedbackNV, &args ))) WARN( "glResumeTransformFeedbackNV returned %#lx\n", status );
 }
 
-static void WINAPI glRotatex( GLfixed angle, GLfixed x, GLfixed y, GLfixed z )
-{
-    struct glRotatex_params args = { .teb = NtCurrentTeb(), .angle = angle, .x = x, .y = y, .z = z };
-    NTSTATUS status;
-    TRACE( "angle %d, x %d, y %d, z %d\n", angle, x, y, z );
-    if ((status = UNIX_CALL( glRotatex, &args ))) WARN( "glRotatex returned %#lx\n", status );
-}
-
 static void WINAPI glRotatexOES( GLfixed angle, GLfixed x, GLfixed y, GLfixed z )
 {
     struct glRotatexOES_params args = { .teb = NtCurrentTeb(), .angle = angle, .x = x, .y = y, .z = z };
@@ -17808,14 +17370,6 @@
     if ((status = UNIX_CALL( glSampleCoverageARB, &args ))) WARN( "glSampleCoverageARB returned %#lx\n", status );
 }
 
-static void WINAPI glSampleCoveragex( GLclampx value, GLboolean invert )
-{
-    struct glSampleCoveragex_params args = { .teb = NtCurrentTeb(), .value = value, .invert = invert };
-    NTSTATUS status;
-    TRACE( "value %d, invert %d\n", value, invert );
-    if ((status = UNIX_CALL( glSampleCoveragex, &args ))) WARN( "glSampleCoveragex returned %#lx\n", status );
-}
-
 static void WINAPI glSampleMapATI( GLuint dst, GLuint interp, GLenum swizzle )
 {
     struct glSampleMapATI_params args = { .teb = NtCurrentTeb(), .dst = dst, .interp = interp, .swizzle = swizzle };
@@ -17920,14 +17474,6 @@
     if ((status = UNIX_CALL( glSamplerParameteriv, &args ))) WARN( "glSamplerParameteriv returned %#lx\n", status );
 }
 
-static void WINAPI glScalex( GLfixed x, GLfixed y, GLfixed z )
-{
-    struct glScalex_params args = { .teb = NtCurrentTeb(), .x = x, .y = y, .z = z };
-    NTSTATUS status;
-    TRACE( "x %d, y %d, z %d\n", x, y, z );
-    if ((status = UNIX_CALL( glScalex, &args ))) WARN( "glScalex returned %#lx\n", status );
-}
-
 static void WINAPI glScalexOES( GLfixed x, GLfixed y, GLfixed z )
 {
     struct glScalexOES_params args = { .teb = NtCurrentTeb(), .x = x, .y = y, .z = z };
@@ -18320,14 +17866,6 @@
     if ((status = UNIX_CALL( glSelectTextureSGIS, &args ))) WARN( "glSelectTextureSGIS returned %#lx\n", status );
 }
 
-static void WINAPI glSemaphoreParameterivNV( GLuint semaphore, GLenum pname, const GLint *params )
-{
-    struct glSemaphoreParameterivNV_params args = { .teb = NtCurrentTeb(), .semaphore = semaphore, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "semaphore %d, pname %d, params %p\n", semaphore, pname, params );
-    if ((status = UNIX_CALL( glSemaphoreParameterivNV, &args ))) WARN( "glSemaphoreParameterivNV returned %#lx\n", status );
-}
-
 static void WINAPI glSemaphoreParameterui64vEXT( GLuint semaphore, GLenum pname, const GLuint64 *params )
 {
     struct glSemaphoreParameterui64vEXT_params args = { .teb = NtCurrentTeb(), .semaphore = semaphore, .pname = pname, .params = params };
@@ -18400,11 +17938,11 @@
     if ((status = UNIX_CALL( glSetMultisamplefvAMD, &args ))) WARN( "glSetMultisamplefvAMD returned %#lx\n", status );
 }
 
-static void WINAPI glShaderBinary( GLsizei count, const GLuint *shaders, GLenum binaryFormat, const void *binary, GLsizei length )
-{
-    struct glShaderBinary_params args = { .teb = NtCurrentTeb(), .count = count, .shaders = shaders, .binaryFormat = binaryFormat, .binary = binary, .length = length };
-    NTSTATUS status;
-    TRACE( "count %d, shaders %p, binaryFormat %d, binary %p, length %d\n", count, shaders, binaryFormat, binary, length );
+static void WINAPI glShaderBinary( GLsizei count, const GLuint *shaders, GLenum binaryformat, const void *binary, GLsizei length )
+{
+    struct glShaderBinary_params args = { .teb = NtCurrentTeb(), .count = count, .shaders = shaders, .binaryformat = binaryformat, .binary = binary, .length = length };
+    NTSTATUS status;
+    TRACE( "count %d, shaders %p, binaryformat %d, binary %p, length %d\n", count, shaders, binaryformat, binary, length );
     if ((status = UNIX_CALL( glShaderBinary, &args ))) WARN( "glShaderBinary returned %#lx\n", status );
 }
 
@@ -18430,22 +17968,6 @@
     NTSTATUS status;
     TRACE( "op %d, res %d, arg1 %d, arg2 %d, arg3 %d\n", op, res, arg1, arg2, arg3 );
     if ((status = UNIX_CALL( glShaderOp3EXT, &args ))) WARN( "glShaderOp3EXT returned %#lx\n", status );
-}
-
-static void WINAPI glShaderSource( GLuint shader, GLsizei count, const GLchar *const*string, const GLint *length )
-{
-    struct glShaderSource_params args = { .teb = NtCurrentTeb(), .shader = shader, .count = count, .string = string, .length = length };
-    NTSTATUS status;
-    TRACE( "shader %d, count %d, string %p, length %p\n", shader, count, string, length );
-    if ((status = UNIX_CALL( glShaderSource, &args ))) WARN( "glShaderSource returned %#lx\n", status );
-}
-
-static void WINAPI glShaderSourceARB( GLhandleARB shaderObj, GLsizei count, const GLcharARB **string, const GLint *length )
-{
-    struct glShaderSourceARB_params args = { .teb = NtCurrentTeb(), .shaderObj = shaderObj, .count = count, .string = string, .length = length };
-    NTSTATUS status;
-    TRACE( "shaderObj %d, count %d, string %p, length %p\n", shaderObj, count, string, length );
-    if ((status = UNIX_CALL( glShaderSourceARB, &args ))) WARN( "glShaderSourceARB returned %#lx\n", status );
 }
 
 static void WINAPI glShaderStorageBlockBinding( GLuint program, GLuint storageBlockIndex, GLuint storageBlockBinding )
@@ -19355,14 +18877,6 @@
     if ((status = UNIX_CALL( glTexCoordPointervINTEL, &args ))) WARN( "glTexCoordPointervINTEL returned %#lx\n", status );
 }
 
-static void WINAPI glTexEnvx( GLenum target, GLenum pname, GLfixed param )
-{
-    struct glTexEnvx_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "target %d, pname %d, param %d\n", target, pname, param );
-    if ((status = UNIX_CALL( glTexEnvx, &args ))) WARN( "glTexEnvx returned %#lx\n", status );
-}
-
 static void WINAPI glTexEnvxOES( GLenum target, GLenum pname, GLfixed param )
 {
     struct glTexEnvxOES_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .param = param };
@@ -19371,14 +18885,6 @@
     if ((status = UNIX_CALL( glTexEnvxOES, &args ))) WARN( "glTexEnvxOES returned %#lx\n", status );
 }
 
-static void WINAPI glTexEnvxv( GLenum target, GLenum pname, const GLfixed *params )
-{
-    struct glTexEnvxv_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "target %d, pname %d, params %p\n", target, pname, params );
-    if ((status = UNIX_CALL( glTexEnvxv, &args ))) WARN( "glTexEnvxv returned %#lx\n", status );
-}
-
 static void WINAPI glTexEnvxvOES( GLenum target, GLenum pname, const GLfixed *params )
 {
     struct glTexEnvxvOES_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .params = params };
@@ -19475,14 +18981,6 @@
     if ((status = UNIX_CALL( glTexPageCommitmentARB, &args ))) WARN( "glTexPageCommitmentARB returned %#lx\n", status );
 }
 
-static void WINAPI glTexPageCommitmentMemNV( GLenum target, GLint layer, GLint level, GLint xoffset, GLint yoffset, GLint zoffset, GLsizei width, GLsizei height, GLsizei depth, GLuint memory, GLuint64 offset, GLboolean commit )
-{
-    struct glTexPageCommitmentMemNV_params args = { .teb = NtCurrentTeb(), .target = target, .layer = layer, .level = level, .xoffset = xoffset, .yoffset = yoffset, .zoffset = zoffset, .width = width, .height = height, .depth = depth, .memory = memory, .offset = offset, .commit = commit };
-    NTSTATUS status;
-    TRACE( "target %d, layer %d, level %d, xoffset %d, yoffset %d, zoffset %d, width %d, height %d, depth %d, memory %d, offset %s, commit %d\n", target, layer, level, xoffset, yoffset, zoffset, width, height, depth, memory, wine_dbgstr_longlong(offset), commit );
-    if ((status = UNIX_CALL( glTexPageCommitmentMemNV, &args ))) WARN( "glTexPageCommitmentMemNV returned %#lx\n", status );
-}
-
 static void WINAPI glTexParameterIiv( GLenum target, GLenum pname, const GLint *params )
 {
     struct glTexParameterIiv_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .params = params };
@@ -19515,14 +19013,6 @@
     if ((status = UNIX_CALL( glTexParameterIuivEXT, &args ))) WARN( "glTexParameterIuivEXT returned %#lx\n", status );
 }
 
-static void WINAPI glTexParameterx( GLenum target, GLenum pname, GLfixed param )
-{
-    struct glTexParameterx_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .param = param };
-    NTSTATUS status;
-    TRACE( "target %d, pname %d, param %d\n", target, pname, param );
-    if ((status = UNIX_CALL( glTexParameterx, &args ))) WARN( "glTexParameterx returned %#lx\n", status );
-}
-
 static void WINAPI glTexParameterxOES( GLenum target, GLenum pname, GLfixed param )
 {
     struct glTexParameterxOES_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .param = param };
@@ -19531,14 +19021,6 @@
     if ((status = UNIX_CALL( glTexParameterxOES, &args ))) WARN( "glTexParameterxOES returned %#lx\n", status );
 }
 
-static void WINAPI glTexParameterxv( GLenum target, GLenum pname, const GLfixed *params )
-{
-    struct glTexParameterxv_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .params = params };
-    NTSTATUS status;
-    TRACE( "target %d, pname %d, params %p\n", target, pname, params );
-    if ((status = UNIX_CALL( glTexParameterxv, &args ))) WARN( "glTexParameterxv returned %#lx\n", status );
-}
-
 static void WINAPI glTexParameterxvOES( GLenum target, GLenum pname, const GLfixed *params )
 {
     struct glTexParameterxvOES_params args = { .teb = NtCurrentTeb(), .target = target, .pname = pname, .params = params };
@@ -19563,14 +19045,6 @@
     if ((status = UNIX_CALL( glTexStorage1D, &args ))) WARN( "glTexStorage1D returned %#lx\n", status );
 }
 
-static void WINAPI glTexStorage1DEXT( GLenum target, GLsizei levels, GLenum internalformat, GLsizei width )
-{
-    struct glTexStorage1DEXT_params args = { .teb = NtCurrentTeb(), .target = target, .levels = levels, .internalformat = internalformat, .width = width };
-    NTSTATUS status;
-    TRACE( "target %d, levels %d, internalformat %d, width %d\n", target, levels, internalformat, width );
-    if ((status = UNIX_CALL( glTexStorage1DEXT, &args ))) WARN( "glTexStorage1DEXT returned %#lx\n", status );
-}
-
 static void WINAPI glTexStorage2D( GLenum target, GLsizei levels, GLenum internalformat, GLsizei width, GLsizei height )
 {
     struct glTexStorage2D_params args = { .teb = NtCurrentTeb(), .target = target, .levels = levels, .internalformat = internalformat, .width = width, .height = height };
@@ -19579,14 +19053,6 @@
     if ((status = UNIX_CALL( glTexStorage2D, &args ))) WARN( "glTexStorage2D returned %#lx\n", status );
 }
 
-static void WINAPI glTexStorage2DEXT( GLenum target, GLsizei levels, GLenum internalformat, GLsizei width, GLsizei height )
-{
-    struct glTexStorage2DEXT_params args = { .teb = NtCurrentTeb(), .target = target, .levels = levels, .internalformat = internalformat, .width = width, .height = height };
-    NTSTATUS status;
-    TRACE( "target %d, levels %d, internalformat %d, width %d, height %d\n", target, levels, internalformat, width, height );
-    if ((status = UNIX_CALL( glTexStorage2DEXT, &args ))) WARN( "glTexStorage2DEXT returned %#lx\n", status );
-}
-
 static void WINAPI glTexStorage2DMultisample( GLenum target, GLsizei samples, GLenum internalformat, GLsizei width, GLsizei height, GLboolean fixedsamplelocations )
 {
     struct glTexStorage2DMultisample_params args = { .teb = NtCurrentTeb(), .target = target, .samples = samples, .internalformat = internalformat, .width = width, .height = height, .fixedsamplelocations = fixedsamplelocations };
@@ -19603,14 +19069,6 @@
     if ((status = UNIX_CALL( glTexStorage3D, &args ))) WARN( "glTexStorage3D returned %#lx\n", status );
 }
 
-static void WINAPI glTexStorage3DEXT( GLenum target, GLsizei levels, GLenum internalformat, GLsizei width, GLsizei height, GLsizei depth )
-{
-    struct glTexStorage3DEXT_params args = { .teb = NtCurrentTeb(), .target = target, .levels = levels, .internalformat = internalformat, .width = width, .height = height, .depth = depth };
-    NTSTATUS status;
-    TRACE( "target %d, levels %d, internalformat %d, width %d, height %d, depth %d\n", target, levels, internalformat, width, height, depth );
-    if ((status = UNIX_CALL( glTexStorage3DEXT, &args ))) WARN( "glTexStorage3DEXT returned %#lx\n", status );
-}
-
 static void WINAPI glTexStorage3DMultisample( GLenum target, GLsizei samples, GLenum internalformat, GLsizei width, GLsizei height, GLsizei depth, GLboolean fixedsamplelocations )
 {
     struct glTexStorage3DMultisample_params args = { .teb = NtCurrentTeb(), .target = target, .samples = samples, .internalformat = internalformat, .width = width, .height = height, .depth = depth, .fixedsamplelocations = fixedsamplelocations };
@@ -19859,14 +19317,6 @@
     if ((status = UNIX_CALL( glTexturePageCommitmentEXT, &args ))) WARN( "glTexturePageCommitmentEXT returned %#lx\n", status );
 }
 
-static void WINAPI glTexturePageCommitmentMemNV( GLuint texture, GLint layer, GLint level, GLint xoffset, GLint yoffset, GLint zoffset, GLsizei width, GLsizei height, GLsizei depth, GLuint memory, GLuint64 offset, GLboolean commit )
-{
-    struct glTexturePageCommitmentMemNV_params args = { .teb = NtCurrentTeb(), .texture = texture, .layer = layer, .level = level, .xoffset = xoffset, .yoffset = yoffset, .zoffset = zoffset, .width = width, .height = height, .depth = depth, .memory = memory, .offset = offset, .commit = commit };
-    NTSTATUS status;
-    TRACE( "texture %d, layer %d, level %d, xoffset %d, yoffset %d, zoffset %d, width %d, height %d, depth %d, memory %d, offset %s, commit %d\n", texture, layer, level, xoffset, yoffset, zoffset, width, height, depth, memory, wine_dbgstr_longlong(offset), commit );
-    if ((status = UNIX_CALL( glTexturePageCommitmentMemNV, &args ))) WARN( "glTexturePageCommitmentMemNV returned %#lx\n", status );
-}
-
 static void WINAPI glTextureParameterIiv( GLuint texture, GLenum pname, const GLint *params )
 {
     struct glTextureParameterIiv_params args = { .teb = NtCurrentTeb(), .texture = texture, .pname = pname, .params = params };
@@ -20235,14 +19685,6 @@
     if ((status = UNIX_CALL( glTransformPathNV, &args ))) WARN( "glTransformPathNV returned %#lx\n", status );
 }
 
-static void WINAPI glTranslatex( GLfixed x, GLfixed y, GLfixed z )
-{
-    struct glTranslatex_params args = { .teb = NtCurrentTeb(), .x = x, .y = y, .z = z };
-    NTSTATUS status;
-    TRACE( "x %d, y %d, z %d\n", x, y, z );
-    if ((status = UNIX_CALL( glTranslatex, &args ))) WARN( "glTranslatex returned %#lx\n", status );
-}
-
 static void WINAPI glTranslatexOES( GLfixed x, GLfixed y, GLfixed z )
 {
     struct glTranslatexOES_params args = { .teb = NtCurrentTeb(), .x = x, .y = y, .z = z };
@@ -21201,42 +20643,6 @@
     NTSTATUS status;
     TRACE( "\n" );
     if ((status = UNIX_CALL( glUnlockArraysEXT, &args ))) WARN( "glUnlockArraysEXT returned %#lx\n", status );
-}
-
-static GLboolean WINAPI glUnmapBuffer( GLenum target )
-{
-    struct glUnmapBuffer_params args = { .teb = NtCurrentTeb(), .target = target };
-    NTSTATUS status;
-    TRACE( "target %d\n", target );
-    if ((status = UNIX_CALL( glUnmapBuffer, &args ))) WARN( "glUnmapBuffer returned %#lx\n", status );
-    return args.ret;
-}
-
-static GLboolean WINAPI glUnmapBufferARB( GLenum target )
-{
-    struct glUnmapBufferARB_params args = { .teb = NtCurrentTeb(), .target = target };
-    NTSTATUS status;
-    TRACE( "target %d\n", target );
-    if ((status = UNIX_CALL( glUnmapBufferARB, &args ))) WARN( "glUnmapBufferARB returned %#lx\n", status );
-    return args.ret;
-}
-
-static GLboolean WINAPI glUnmapNamedBuffer( GLuint buffer )
-{
-    struct glUnmapNamedBuffer_params args = { .teb = NtCurrentTeb(), .buffer = buffer };
-    NTSTATUS status;
-    TRACE( "buffer %d\n", buffer );
-    if ((status = UNIX_CALL( glUnmapNamedBuffer, &args ))) WARN( "glUnmapNamedBuffer returned %#lx\n", status );
-    return args.ret;
-}
-
-static GLboolean WINAPI glUnmapNamedBufferEXT( GLuint buffer )
-{
-    struct glUnmapNamedBufferEXT_params args = { .teb = NtCurrentTeb(), .buffer = buffer };
-    NTSTATUS status;
-    TRACE( "buffer %d\n", buffer );
-    if ((status = UNIX_CALL( glUnmapNamedBufferEXT, &args ))) WARN( "glUnmapNamedBufferEXT returned %#lx\n", status );
-    return args.ret;
 }
 
 static void WINAPI glUnmapObjectBufferATI( GLuint buffer )
@@ -24877,6 +24283,19 @@
 }
 
 extern const GLubyte * WINAPI glGetStringi( GLenum name, GLuint index );
+extern void * WINAPI glMapBuffer( GLenum target, GLenum access );
+extern void * WINAPI glMapBufferARB( GLenum target, GLenum access );
+extern void * WINAPI glMapBufferRange( GLenum target, GLintptr offset, GLsizeiptr length, GLbitfield access );
+extern void * WINAPI glMapNamedBuffer( GLuint buffer, GLenum access );
+extern void * WINAPI glMapNamedBufferEXT( GLuint buffer, GLenum access );
+extern void * WINAPI glMapNamedBufferRange( GLuint buffer, GLintptr offset, GLsizeiptr length, GLbitfield access );
+extern void * WINAPI glMapNamedBufferRangeEXT( GLuint buffer, GLintptr offset, GLsizeiptr length, GLbitfield access );
+extern void WINAPI glShaderSource( GLuint shader, GLsizei count, const GLchar *const*string, const GLint *length );
+extern void WINAPI glShaderSourceARB( GLhandleARB shaderObj, GLsizei count, const GLcharARB **string, const GLint *length );
+extern GLboolean WINAPI glUnmapBuffer( GLenum target );
+extern GLboolean WINAPI glUnmapBufferARB( GLenum target );
+extern GLboolean WINAPI glUnmapNamedBuffer( GLuint buffer );
+extern GLboolean WINAPI glUnmapNamedBufferEXT( GLuint buffer );
 extern BOOL WINAPI wglChoosePixelFormatARB( HDC hdc, const int *piAttribIList, const FLOAT *pfAttribFList, UINT nMaxFormats, int *piFormats, UINT *nNumFormats );
 extern HDC WINAPI wglGetCurrentReadDCARB(void);
 extern const char * WINAPI wglGetExtensionsStringARB( HDC hdc );
@@ -24898,7 +24317,6 @@
     glAlphaFragmentOp1ATI,
     glAlphaFragmentOp2ATI,
     glAlphaFragmentOp3ATI,
-    glAlphaFuncx,
     glAlphaFuncxOES,
     glAlphaToCoverageDitherControlNV,
     glApplyFramebufferAttachmentCMAAINTEL,
@@ -25015,15 +24433,12 @@
     glBlendParameteriNV,
     glBlitFramebuffer,
     glBlitFramebufferEXT,
-    glBlitFramebufferLayerEXT,
-    glBlitFramebufferLayersEXT,
     glBlitNamedFramebuffer,
     glBufferAddressRangeNV,
     glBufferAttachMemoryNV,
     glBufferData,
     glBufferDataARB,
     glBufferPageCommitmentARB,
-    glBufferPageCommitmentMemNV,
     glBufferParameteriAPPLE,
     glBufferRegionEnabled,
     glBufferStorage,
@@ -25047,12 +24462,10 @@
     glClearBufferuiv,
     glClearColorIiEXT,
     glClearColorIuiEXT,
-    glClearColorx,
     glClearColorxOES,
     glClearDepthdNV,
     glClearDepthf,
     glClearDepthfOES,
-    glClearDepthx,
     glClearDepthxOES,
     glClearNamedBufferData,
     glClearNamedBufferDataEXT,
@@ -25071,9 +24484,7 @@
     glClientWaitSemaphoreui64NVX,
     glClientWaitSync,
     glClipControl,
-    glClipPlanef,
     glClipPlanefOES,
-    glClipPlanex,
     glClipPlanexOES,
     glColor3fVertex3fSUN,
     glColor3fVertex3fvSUN,
@@ -25089,7 +24500,6 @@
     glColor4ubVertex2fvSUN,
     glColor4ubVertex3fSUN,
     glColor4ubVertex3fvSUN,
-    glColor4x,
     glColor4xOES,
     glColor4xvOES,
     glColorFormatNV,
@@ -25219,7 +24629,6 @@
     glCreateQueries,
     glCreateRenderbuffers,
     glCreateSamplers,
-    glCreateSemaphoresNV,
     glCreateShader,
     glCreateShaderObjectARB,
     glCreateShaderProgramEXT,
@@ -25292,7 +24701,6 @@
     glDepthRangedNV,
     glDepthRangef,
     glDepthRangefOES,
-    glDepthRangex,
     glDepthRangexOES,
     glDetachObjectARB,
     glDetachShader,
@@ -25336,8 +24744,6 @@
     glDrawElementsInstancedBaseVertexBaseInstance,
     glDrawElementsInstancedEXT,
     glDrawMeshArraysSUN,
-    glDrawMeshTasksEXT,
-    glDrawMeshTasksIndirectEXT,
     glDrawMeshTasksIndirectNV,
     glDrawMeshTasksNV,
     glDrawRangeElementArrayAPPLE,
@@ -25425,9 +24831,7 @@
     glFogCoordhNV,
     glFogCoordhvNV,
     glFogFuncSGIS,
-    glFogx,
     glFogxOES,
-    glFogxv,
     glFogxvOES,
     glFragmentColorMaterialSGIX,
     glFragmentCoverageColorNV,
@@ -25472,9 +24876,7 @@
     glFramebufferTextureLayerEXT,
     glFramebufferTextureMultiviewOVR,
     glFreeObjectBufferATI,
-    glFrustumf,
     glFrustumfOES,
-    glFrustumx,
     glFrustumxOES,
     glGenAsyncMarkersSGIX,
     glGenBuffers,
@@ -25539,9 +24941,7 @@
     glGetBufferPointervARB,
     glGetBufferSubData,
     glGetBufferSubDataARB,
-    glGetClipPlanef,
     glGetClipPlanefOES,
-    glGetClipPlanex,
     glGetClipPlanexOES,
     glGetColorTable,
     glGetColorTableEXT,
@@ -25583,7 +24983,6 @@
     glGetFinalCombinerInputParameterfvNV,
     glGetFinalCombinerInputParameterivNV,
     glGetFirstPerfQueryIdINTEL,
-    glGetFixedv,
     glGetFixedvOES,
     glGetFloatIndexedvEXT,
     glGetFloati_v,
@@ -25631,7 +25030,6 @@
     glGetInvariantFloatvEXT,
     glGetInvariantIntegervEXT,
     glGetLightxOES,
-    glGetLightxv,
     glGetListParameterfvSGIX,
     glGetListParameterivSGIX,
     glGetLocalConstantBooleanvEXT,
@@ -25644,7 +25042,6 @@
     glGetMapParameterivNV,
     glGetMapxvOES,
     glGetMaterialxOES,
-    glGetMaterialxv,
     glGetMemoryObjectDetachedResourcesuivNV,
     glGetMemoryObjectParameterivEXT,
     glGetMinmax,
@@ -25783,7 +25180,6 @@
     glGetSamplerParameterIuiv,
     glGetSamplerParameterfv,
     glGetSamplerParameteriv,
-    glGetSemaphoreParameterivNV,
     glGetSemaphoreParameterui64vEXT,
     glGetSeparableFilter,
     glGetSeparableFilterEXT,
@@ -25802,7 +25198,6 @@
     glGetSynciv,
     glGetTexBumpParameterfvATI,
     glGetTexBumpParameterivATI,
-    glGetTexEnvxv,
     glGetTexEnvxvOES,
     glGetTexFilterFuncSGIS,
     glGetTexGenxvOES,
@@ -25812,7 +25207,6 @@
     glGetTexParameterIuiv,
     glGetTexParameterIuivEXT,
     glGetTexParameterPointervAPPLE,
-    glGetTexParameterxv,
     glGetTexParameterxvOES,
     glGetTextureHandleARB,
     glGetTextureHandleNV,
@@ -26035,15 +25429,10 @@
     glLGPUNamedBufferSubDataNVX,
     glLabelObjectEXT,
     glLightEnviSGIX,
-    glLightModelx,
     glLightModelxOES,
-    glLightModelxv,
     glLightModelxvOES,
-    glLightx,
     glLightxOES,
-    glLightxv,
     glLightxvOES,
-    glLineWidthx,
     glLineWidthxOES,
     glLinkProgram,
     glLinkProgramARB,
@@ -26053,7 +25442,6 @@
     glListParameteriSGIX,
     glListParameterivSGIX,
     glLoadIdentityDeformationMapSGIX,
-    glLoadMatrixx,
     glLoadMatrixxOES,
     glLoadProgramNV,
     glLoadTransposeMatrixd,
@@ -26096,9 +25484,7 @@
     glMapVertexAttrib1fAPPLE,
     glMapVertexAttrib2dAPPLE,
     glMapVertexAttrib2fAPPLE,
-    glMaterialx,
     glMaterialxOES,
-    glMaterialxv,
     glMaterialxvOES,
     glMatrixFrustumEXT,
     glMatrixIndexPointerARB,
@@ -26139,7 +25525,6 @@
     glMinSampleShadingARB,
     glMinmax,
     glMinmaxEXT,
-    glMultMatrixx,
     glMultMatrixxOES,
     glMultTransposeMatrixd,
     glMultTransposeMatrixdARB,
@@ -26164,9 +25549,7 @@
     glMultiDrawElementsIndirectBindlessNV,
     glMultiDrawElementsIndirectCount,
     glMultiDrawElementsIndirectCountARB,
-    glMultiDrawMeshTasksIndirectCountEXT,
     glMultiDrawMeshTasksIndirectCountNV,
-    glMultiDrawMeshTasksIndirectEXT,
     glMultiDrawMeshTasksIndirectNV,
     glMultiDrawRangeElementArrayAPPLE,
     glMultiModeDrawArraysIBM,
@@ -26290,7 +25673,6 @@
     glMultiTexCoord4sv,
     glMultiTexCoord4svARB,
     glMultiTexCoord4svSGIS,
-    glMultiTexCoord4x,
     glMultiTexCoord4xOES,
     glMultiTexCoord4xvOES,
     glMultiTexCoordP1ui,
@@ -26345,7 +25727,6 @@
     glNamedBufferDataEXT,
     glNamedBufferPageCommitmentARB,
     glNamedBufferPageCommitmentEXT,
-    glNamedBufferPageCommitmentMemNV,
     glNamedBufferStorage,
     glNamedBufferStorageEXT,
     glNamedBufferStorageExternalEXT,
@@ -26371,7 +25752,6 @@
     glNamedFramebufferTextureFaceEXT,
     glNamedFramebufferTextureLayer,
     glNamedFramebufferTextureLayerEXT,
-    glNamedFramebufferTextureMultiviewOVR,
     glNamedProgramLocalParameter4dEXT,
     glNamedProgramLocalParameter4dvEXT,
     glNamedProgramLocalParameter4fEXT,
@@ -26397,7 +25777,6 @@
     glNormal3fVertex3fvSUN,
     glNormal3hNV,
     glNormal3hvNV,
-    glNormal3x,
     glNormal3xOES,
     glNormal3xvOES,
     glNormalFormatNV,
@@ -26420,9 +25799,7 @@
     glObjectPtrLabel,
     glObjectPurgeableAPPLE,
     glObjectUnpurgeableAPPLE,
-    glOrthof,
     glOrthofOES,
-    glOrthox,
     glOrthoxOES,
     glPNTrianglesfATI,
     glPNTrianglesiATI,
@@ -26480,17 +25857,13 @@
     glPointParameteriNV,
     glPointParameteriv,
     glPointParameterivNV,
-    glPointParameterx,
-    glPointParameterxv,
     glPointParameterxvOES,
-    glPointSizex,
     glPointSizexOES,
     glPollAsyncSGIX,
     glPollInstrumentsSGIX,
     glPolygonOffsetClamp,
     glPolygonOffsetClampEXT,
     glPolygonOffsetEXT,
-    glPolygonOffsetx,
     glPolygonOffsetxOES,
     glPopDebugGroup,
     glPopGroupMarkerEXT,
@@ -26749,11 +26122,9 @@
     glResolveDepthValuesNV,
     glResumeTransformFeedback,
     glResumeTransformFeedbackNV,
-    glRotatex,
     glRotatexOES,
     glSampleCoverage,
     glSampleCoverageARB,
-    glSampleCoveragex,
     glSampleMapATI,
     glSampleMaskEXT,
     glSampleMaskIndexedNV,
@@ -26767,7 +26138,6 @@
     glSamplerParameterfv,
     glSamplerParameteri,
     glSamplerParameteriv,
-    glScalex,
     glScalexOES,
     glScissorArrayv,
     glScissorExclusiveArrayvNV,
@@ -26817,7 +26187,6 @@
     glSelectPerfMonitorCountersAMD,
     glSelectTextureCoordSetSGIS,
     glSelectTextureSGIS,
-    glSemaphoreParameterivNV,
     glSemaphoreParameterui64vEXT,
     glSeparableFilter2D,
     glSeparableFilter2DEXT,
@@ -26946,9 +26315,7 @@
     glTexCoordPointerEXT,
     glTexCoordPointerListIBM,
     glTexCoordPointervINTEL,
-    glTexEnvx,
     glTexEnvxOES,
-    glTexEnvxv,
     glTexEnvxvOES,
     glTexFilterFuncSGIS,
     glTexGenxOES,
@@ -26961,23 +26328,17 @@
     glTexImage3DMultisampleCoverageNV,
     glTexImage4DSGIS,
     glTexPageCommitmentARB,
-    glTexPageCommitmentMemNV,
     glTexParameterIiv,
     glTexParameterIivEXT,
     glTexParameterIuiv,
     glTexParameterIuivEXT,
-    glTexParameterx,
     glTexParameterxOES,
-    glTexParameterxv,
     glTexParameterxvOES,
     glTexRenderbufferNV,
     glTexStorage1D,
-    glTexStorage1DEXT,
     glTexStorage2D,
-    glTexStorage2DEXT,
     glTexStorage2DMultisample,
     glTexStorage3D,
-    glTexStorage3DEXT,
     glTexStorage3DMultisample,
     glTexStorageMem1DEXT,
     glTexStorageMem2DEXT,
@@ -27009,7 +26370,6 @@
     glTextureMaterialEXT,
     glTextureNormalEXT,
     glTexturePageCommitmentEXT,
-    glTexturePageCommitmentMemNV,
     glTextureParameterIiv,
     glTextureParameterIivEXT,
     glTextureParameterIuiv,
@@ -27056,7 +26416,6 @@
     glTransformFeedbackVaryingsEXT,
     glTransformFeedbackVaryingsNV,
     glTransformPathNV,
-    glTranslatex,
     glTranslatexOES,
     glUniform1d,
     glUniform1dv,
