--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -64,7 +64,6 @@
 static const char devpropkey_monitor_gpu_luidA[] = "Properties\\{CA085853-16CE-48AA-B114-DE9C72334223}\\0001";
 static const char devpropkey_monitor_output_idA[] = "Properties\\{CA085853-16CE-48AA-B114-DE9C72334223}\\0002";
 static const char wine_devpropkey_monitor_rcworkA[] = "Properties\\{233a9ef3-afc4-4abd-b564-c32f21f1535b}\\0004";
-static const char wine_devpropkey_monitor_hdr_enabledA[] = "Properties\\{233a9ef3-afc4-4abd-b564-c32f21f1535b}\\0006";
 
 static const WCHAR linkedW[] = {'L','i','n','k','e','d',0};
 static const WCHAR symbolic_link_valueW[] =
@@ -156,7 +155,6 @@
     RECT rc_work;
     BOOL is_clone;
     struct edid_monitor_info edid_info;
-    BOOL hdr_enabled;
 };
 
 static struct list gpus = LIST_INIT(gpus);
@@ -166,7 +164,9 @@
 static UINT64 monitor_update_serial;
 static pthread_mutex_t display_lock = PTHREAD_MUTEX_INITIALIZER;
 
-static BOOL emulate_modeset;
+static BOOL emulate_modelist = TRUE;
+static BOOL emulate_modeset = TRUE;
+static UINT limit_resolutions = 0;
 BOOL decorated_mode = TRUE;
 UINT64 thunk_lock_callback = 0;
 
@@ -757,16 +757,6 @@
         NtClose( subkey );
     }
 
-    /* WINE_DEVPROPKEY_MONITOR_HDR_ENABLED */
-    size = query_reg_subkey_value( hkey, wine_devpropkey_monitor_hdr_enabledA,
-                                   value, sizeof(buffer) );
-    if (size != sizeof(monitor->hdr_enabled))
-    {
-        NtClose( hkey );
-        return FALSE;
-    }
-    monitor->hdr_enabled = *(const BOOL *)value->Data;
-
     NtClose( hkey );
     return TRUE;
 }
@@ -787,7 +777,7 @@
     return TRUE;
 }
 
-static void reg_empty_key( HKEY root, const char *key_name, BOOL subkeys_only )
+static void reg_empty_key( HKEY root, const char *key_name )
 {
     char buffer[4096];
     KEY_NODE_INFORMATION *key = (KEY_NODE_INFORMATION *)buffer;
@@ -798,13 +788,10 @@
     while (!NtEnumerateKey( hkey, 0, KeyNodeInformation, key, sizeof(buffer), &size ))
         reg_delete_tree( hkey, key->Name, key->NameLength );
 
-    if (!subkeys_only)
+    while (!NtEnumerateValueKey( hkey, 0, KeyValueFullInformation, value, sizeof(buffer), &size ))
     {
-        while (!NtEnumerateValueKey( hkey, 0, KeyValueFullInformation, value, sizeof(buffer), &size ))
-        {
-            UNICODE_STRING name = { value->NameLength, value->NameLength, value->Name };
-            NtDeleteValueKey( hkey, &name );
-        }
+        UNICODE_STRING name = { value->NameLength, value->NameLength, value->Name };
+        NtDeleteValueKey( hkey, &name );
     }
 
     if (hkey != root) NtClose( hkey );
@@ -861,29 +848,24 @@
     if (!video_key) video_key = reg_create_ascii_key( NULL, devicemap_video_keyA, REG_OPTION_VOLATILE, NULL );
 
     /* delete monitors */
-    reg_empty_key( enum_key, "DISPLAY", FALSE );
+    reg_empty_key( enum_key, "DISPLAY" );
     snprintf( buffer, sizeof(buffer), "Class\\%s", guid_devclass_monitorA );
     hkey = reg_create_ascii_key( control_key, buffer, 0, NULL );
-    reg_empty_key( hkey, NULL, FALSE );
+    reg_empty_key( hkey, NULL );
     set_reg_ascii_value( hkey, "", "Monitors" );
     set_reg_ascii_value( hkey, "Class", "Monitor" );
     NtClose( hkey );
 
     /* delete sources */
-    reg_empty_key( video_key, NULL, FALSE );
+    reg_empty_key( video_key, NULL );
 
     /* clean GPUs */
     snprintf( buffer, sizeof(buffer), "Class\\%s", guid_devclass_displayA );
     hkey = reg_create_ascii_key( control_key, buffer, 0, NULL );
-    reg_empty_key( hkey, NULL, FALSE );
+    reg_empty_key( hkey, NULL );
     set_reg_ascii_value( hkey, "", "Display adapters" );
     set_reg_ascii_value( hkey, "Class", "Display" );
     NtClose( hkey );
-    if ((hkey = reg_create_ascii_key( NULL, directx_keyA, 0, NULL )))
-    {
-        reg_empty_key( hkey, NULL, TRUE );
-        NtClose( hkey );
-    }
 
     hkey = reg_open_ascii_key( enum_key, "PCI" );
 
@@ -1113,10 +1095,10 @@
     /* The last seven digits are the driver number. */
     switch (vendor)
     {
-    case 0x8086: /* Intel */    return "35.0.101.6314";
-    case 0x1002: /* AMD */      return "35.0.21025.1024";
-    case 0x10de: /* Nvidia */   return "35.0.15.6094";
-    default:                    return "35.0.10.1000";
+    case 0x8086: /* Intel */    return "35.0.999.9999";
+    case 0x1002: /* AMD */      return "35.0.99999.9999";
+    case 0x10de: /* Nvidia */   return "35.0.99.9999";
+    default:                    return "35.0.99.9999";
     }
 }
 
@@ -1131,355 +1113,20 @@
     }
 }
 
-const char *gpu_device_name( UINT16 vendor, UINT16 device, const char *default_name )
-{
-    switch (MAKELONG(vendor, device))
-    {
-    /* Nvidia cards */
-    case MAKELONG(0x10de, 0x0018): return "NVIDIA RIVA 128";
-    case MAKELONG(0x10de, 0x0020): return "NVIDIA RIVA TNT";
-    case MAKELONG(0x10de, 0x0028): return "NVIDIA RIVA TNT2/TNT2 Pro";
-    case MAKELONG(0x10de, 0x0100): return "NVIDIA GeForce 256";
-    case MAKELONG(0x10de, 0x0150): return "NVIDIA GeForce2 GTS/GeForce2 Pro";
-    case MAKELONG(0x10de, 0x0110): return "NVIDIA GeForce2 MX/MX 400";
-    case MAKELONG(0x10de, 0x0200): return "NVIDIA GeForce3";
-    case MAKELONG(0x10de, 0x0170): return "NVIDIA GeForce4 MX 460";
-    case MAKELONG(0x10de, 0x0253): return "NVIDIA GeForce4 Ti 4200";
-    case MAKELONG(0x10de, 0x0320): return "NVIDIA GeForce FX 5200";
-    case MAKELONG(0x10de, 0x0312): return "NVIDIA GeForce FX 5600";
-    case MAKELONG(0x10de, 0x0302): return "NVIDIA GeForce FX 5800";
-    case MAKELONG(0x10de, 0x014f): return "NVIDIA GeForce 6200";
-    case MAKELONG(0x10de, 0x0140): return "NVIDIA GeForce 6600 GT";
-    case MAKELONG(0x10de, 0x0041): return "NVIDIA GeForce 6800";
-    case MAKELONG(0x10de, 0x01d7): return "NVIDIA GeForce Go 7300";
-    case MAKELONG(0x10de, 0x01d8): return "NVIDIA GeForce Go 7400";
-    case MAKELONG(0x10de, 0x0391): return "NVIDIA GeForce 7600 GT";
-    case MAKELONG(0x10de, 0x0092): return "NVIDIA GeForce 7800 GT";
-    case MAKELONG(0x10de, 0x0849): return "NVIDIA GeForce 8200";
-    case MAKELONG(0x10de, 0x084b): return "NVIDIA GeForce 8200";
-    case MAKELONG(0x10de, 0x0423): return "NVIDIA GeForce 8300 GS";
-    case MAKELONG(0x10de, 0x0404): return "NVIDIA GeForce 8400 GS";
-    case MAKELONG(0x10de, 0x0421): return "NVIDIA GeForce 8500 GT";
-    case MAKELONG(0x10de, 0x0402): return "NVIDIA GeForce 8600 GT";
-    case MAKELONG(0x10de, 0x0407): return "NVIDIA GeForce 8600M GT";
-    case MAKELONG(0x10de, 0x0193): return "NVIDIA GeForce 8800 GTS";
-    case MAKELONG(0x10de, 0x0191): return "NVIDIA GeForce 8800 GTX";
-    case MAKELONG(0x10de, 0x086d): return "NVIDIA GeForce 9200";
-    case MAKELONG(0x10de, 0x086c): return "NVIDIA GeForce 9300";
-    case MAKELONG(0x10de, 0x0863): return "NVIDIA GeForce 9400M";
-    case MAKELONG(0x10de, 0x042c): return "NVIDIA GeForce 9400 GT";
-    case MAKELONG(0x10de, 0x0640): return "NVIDIA GeForce 9500 GT";
-    case MAKELONG(0x10de, 0x0622): return "NVIDIA GeForce 9600 GT";
-    case MAKELONG(0x10de, 0x064a): return "NVIDIA GeForce 9700M GT";
-    case MAKELONG(0x10de, 0x0614): return "NVIDIA GeForce 9800 GT";
-    case MAKELONG(0x10de, 0x0a23): return "NVIDIA GeForce 210";
-    case MAKELONG(0x10de, 0x0a20): return "NVIDIA GeForce GT 220";
-    case MAKELONG(0x10de, 0x0ca3): return "NVIDIA GeForce GT 240";
-    case MAKELONG(0x10de, 0x0615): return "NVIDIA GeForce GTS 250";
-    case MAKELONG(0x10de, 0x05e2): return "NVIDIA GeForce GTX 260";
-    case MAKELONG(0x10de, 0x05e6): return "NVIDIA GeForce GTX 275";
-    case MAKELONG(0x10de, 0x05e1): return "NVIDIA GeForce GTX 280";
-    case MAKELONG(0x10de, 0x0a7a): return "NVIDIA GeForce 315M";
-    case MAKELONG(0x10de, 0x08a3): return "NVIDIA GeForce 320M";
-    case MAKELONG(0x10de, 0x0a2d): return "NVIDIA GeForce GT 320M";
-    case MAKELONG(0x10de, 0x0a35): return "NVIDIA GeForce GT 325M";
-    case MAKELONG(0x10de, 0x0ca0): return "NVIDIA GeForce GT 330";
-    case MAKELONG(0x10de, 0x0cb0): return "NVIDIA GeForce GTS 350M";
-    case MAKELONG(0x10de, 0x1055): return "NVIDIA GeForce 410M";
-    case MAKELONG(0x10de, 0x0de2): return "NVIDIA GeForce GT 420";
-    case MAKELONG(0x10de, 0x0df0): return "NVIDIA GeForce GT 425M";
-    case MAKELONG(0x10de, 0x0de1): return "NVIDIA GeForce GT 430";
-    case MAKELONG(0x10de, 0x0de0): return "NVIDIA GeForce GT 440";
-    case MAKELONG(0x10de, 0x0dc4): return "NVIDIA GeForce GTS 450";
-    case MAKELONG(0x10de, 0x0e22): return "NVIDIA GeForce GTX 460";
-    case MAKELONG(0x10de, 0x0dd1): return "NVIDIA GeForce GTX 460M";
-    case MAKELONG(0x10de, 0x06c4): return "NVIDIA GeForce GTX 465";
-    case MAKELONG(0x10de, 0x06cd): return "NVIDIA GeForce GTX 470";
-    case MAKELONG(0x10de, 0x06c0): return "NVIDIA GeForce GTX 480";
-    case MAKELONG(0x10de, 0x1040): return "NVIDIA GeForce GT 520";
-    case MAKELONG(0x10de, 0x0dec): return "NVIDIA GeForce GT 525M";
-    case MAKELONG(0x10de, 0x0df4): return "NVIDIA GeForce GT 540M";
-    case MAKELONG(0x10de, 0x1244): return "NVIDIA GeForce GTX 550 Ti";
-    case MAKELONG(0x10de, 0x04b8): return "NVIDIA GeForce GT 555M";
-    case MAKELONG(0x10de, 0x1200): return "NVIDIA GeForce GTX 560 Ti";
-    case MAKELONG(0x10de, 0x1251): return "NVIDIA GeForce GTX 560M";
-    case MAKELONG(0x10de, 0x1201): return "NVIDIA GeForce GTX 560";
-    case MAKELONG(0x10de, 0x1081): return "NVIDIA GeForce GTX 570";
-    case MAKELONG(0x10de, 0x1080): return "NVIDIA GeForce GTX 580";
-    case MAKELONG(0x10de, 0x104a): return "NVIDIA GeForce GT 610";
-    case MAKELONG(0x10de, 0x0f00): return "NVIDIA GeForce GT 630";
-    case MAKELONG(0x10de, 0x0de9): return "NVIDIA GeForce GT 630M";
-    case MAKELONG(0x10de, 0x0fc1): return "NVIDIA GeForce GT 640";
-    case MAKELONG(0x10de, 0x0fd2): return "NVIDIA GeForce GT 640M";
-    case MAKELONG(0x10de, 0x0fd1): return "NVIDIA GeForce GT 650M";
-    case MAKELONG(0x10de, 0x0fc6): return "NVIDIA GeForce GTX 650";
-    case MAKELONG(0x10de, 0x11c6): return "NVIDIA GeForce GTX 650 Ti";
-    case MAKELONG(0x10de, 0x11c0): return "NVIDIA GeForce GTX 660";
-    case MAKELONG(0x10de, 0x0fd4): return "NVIDIA GeForce GTX 660M";
-    case MAKELONG(0x10de, 0x1183): return "NVIDIA GeForce GTX 660 Ti";
-    case MAKELONG(0x10de, 0x1189): return "NVIDIA GeForce GTX 670";
-    case MAKELONG(0x10de, 0x11a1): return "NVIDIA GeForce GTX 670MX";
-    case MAKELONG(0x10de, 0x11a7): return "NVIDIA GeForce GTX 675MX";
-    case MAKELONG(0x10de, 0x11a2): return "NVIDIA GeForce GTX 675MX";
-    case MAKELONG(0x10de, 0x1180): return "NVIDIA GeForce GTX 680";
-    case MAKELONG(0x10de, 0x1188): return "NVIDIA GeForce GTX 690";
-    case MAKELONG(0x10de, 0x128b): return "NVIDIA GeForce GT 720";
-    case MAKELONG(0x10de, 0x1287): return "NVIDIA GeForce GT 730";
-    case MAKELONG(0x10de, 0x0fe1): return "NVIDIA GeForce GT 730M";
-    case MAKELONG(0x10de, 0x1292): return "NVIDIA GeForce GT 740M";
-    case MAKELONG(0x10de, 0x0fe9): return "NVIDIA GeForce GT 750M";
-    case MAKELONG(0x10de, 0x0fcd): return "NVIDIA GeForce GT 755M";
-    case MAKELONG(0x10de, 0x1381): return "NVIDIA GeForce GTX 750";
-    case MAKELONG(0x10de, 0x1380): return "NVIDIA GeForce GTX 750 Ti";
-    case MAKELONG(0x10de, 0x1187): return "NVIDIA GeForce GTX 760";
-    case MAKELONG(0x10de, 0x1193): return "NVIDIA GeForce GTX 760 Ti";
-    case MAKELONG(0x10de, 0x11e2): return "NVIDIA GeForce GTX 765M";
-    case MAKELONG(0x10de, 0x11e0): return "NVIDIA GeForce GTX 770M";
-    case MAKELONG(0x10de, 0x1184): return "NVIDIA GeForce GTX 770";
-    case MAKELONG(0x10de, 0x119d): return "NVIDIA GeForce GTX 775M";
-    case MAKELONG(0x10de, 0x1004): return "NVIDIA GeForce GTX 780";
-    case MAKELONG(0x10de, 0x119e): return "NVIDIA GeForce GTX 780M";
-    case MAKELONG(0x10de, 0x100a): return "NVIDIA GeForce GTX 780 Ti";
-    case MAKELONG(0x10de, 0x1005): return "NVIDIA GeForce GTX TITAN";
-    case MAKELONG(0x10de, 0x100c): return "NVIDIA GeForce GTX TITAN Black";
-    case MAKELONG(0x10de, 0x17c2): return "NVIDIA GeForce GTX TITAN X";
-    case MAKELONG(0x10de, 0x1001): return "NVIDIA GeForce GTX TITAN Z";
-    case MAKELONG(0x10de, 0x0fed): return "NVIDIA GeForce 820M";
-    case MAKELONG(0x10de, 0x1340): return "NVIDIA GeForce 830M";
-    case MAKELONG(0x10de, 0x1341): return "NVIDIA GeForce 840M";
-    case MAKELONG(0x10de, 0x1344): return "NVIDIA GeForce 845M";
-    case MAKELONG(0x10de, 0x1391): return "NVIDIA GeForce GTX 850M";
-    case MAKELONG(0x10de, 0x1392): return "NVIDIA GeForce GTX 860M";
-    case MAKELONG(0x10de, 0x119a): return "NVIDIA GeForce GTX 860M";
-    case MAKELONG(0x10de, 0x1199): return "NVIDIA GeForce GTX 870M";
-    case MAKELONG(0x10de, 0x1198): return "NVIDIA GeForce GTX 880M";
-    case MAKELONG(0x10de, 0x1347): return "NVIDIA GeForce 940M";
-    case MAKELONG(0x10de, 0x1402): return "NVIDIA GeForce GTX 950";
-    case MAKELONG(0x10de, 0x139a): return "NVIDIA GeForce GTX 950M";
-    case MAKELONG(0x10de, 0x1401): return "NVIDIA GeForce GTX 960";
-    case MAKELONG(0x10de, 0x139b): return "NVIDIA GeForce GTX 960M";
-    case MAKELONG(0x10de, 0x13c2): return "NVIDIA GeForce GTX 970";
-    case MAKELONG(0x10de, 0x13d8): return "NVIDIA GeForce GTX 970M";
-    case MAKELONG(0x10de, 0x13c0): return "NVIDIA GeForce GTX 980";
-    case MAKELONG(0x10de, 0x17c8): return "NVIDIA GeForce GTX 980 Ti";
-    case MAKELONG(0x10de, 0x1d01): return "NVIDIA GeForce GT 1030";
-    case MAKELONG(0x10de, 0x1c81): return "NVIDIA GeForce GTX 1050";
-    case MAKELONG(0x10de, 0x1c82): return "NVIDIA GeForce GTX 1050 Ti";
-    case MAKELONG(0x10de, 0x1c02): return "NVIDIA GeForce GTX 1060 3GB";
-    case MAKELONG(0x10de, 0x1c03): return "NVIDIA GeForce GTX 1060";
-    case MAKELONG(0x10de, 0x1c20): return "NVIDIA GeForce GTX 1060M";
-    case MAKELONG(0x10de, 0x1b81): return "NVIDIA GeForce GTX 1070";
-    case MAKELONG(0x10de, 0x1be1): return "NVIDIA GeForce GTX 1070M";
-    case MAKELONG(0x10de, 0x1b80): return "NVIDIA GeForce GTX 1080";
-    case MAKELONG(0x10de, 0x1be0): return "NVIDIA GeForce GTX 1080M";
-    case MAKELONG(0x10de, 0x1b06): return "NVIDIA GeForce GTX 1080 Ti";
-    case MAKELONG(0x10de, 0x1b00): return "NVIDIA TITAN X (Pascal)";
-    case MAKELONG(0x10de, 0x1d81): return "NVIDIA TITAN V";
-    case MAKELONG(0x10de, 0x1f82): return "NVIDIA GeForce GTX 1650";
-    case MAKELONG(0x10de, 0x2187): return "NVIDIA GeForce GTX 1650 SUPER";
-    case MAKELONG(0x10de, 0x21c4): return "NVIDIA GeForce GTX 1660 SUPER";
-    case MAKELONG(0x10de, 0x2182): return "NVIDIA GeForce GTX 1660 Ti";
-    case MAKELONG(0x10de, 0x1f08): return "NVIDIA GeForce RTX 2060";
-    case MAKELONG(0x10de, 0x1f07): return "NVIDIA GeForce RTX 2070";
-    case MAKELONG(0x10de, 0x1e87): return "NVIDIA GeForce RTX 2080";
-    case MAKELONG(0x10de, 0x1e07): return "NVIDIA GeForce RTX 2080 Ti";
-    case MAKELONG(0x10de, 0x2507): return "NVIDIA GeForce RTX 3050";
-    case MAKELONG(0x10de, 0x2544): return "NVIDIA GeForce RTX 3060";
-    case MAKELONG(0x10de, 0x2504): return "NVIDIA GeForce RTX 3060 (Low Hash Rate)";
-    case MAKELONG(0x10de, 0x2414): return "NVIDIA GeForce RTX 3060 Ti (GA103)";
-    case MAKELONG(0x10de, 0x2486): return "NVIDIA GeForce RTX 3060 Ti (GA104)";
-    case MAKELONG(0x10de, 0x2489): return "NVIDIA GeForce RTX 3060 Ti (GA104, Low Hash Rate)";
-    case MAKELONG(0x10de, 0x2484): return "NVIDIA GeForce RTX 3070";
-    case MAKELONG(0x10de, 0x2488): return "NVIDIA GeForce RTX 3070 (Low Hash Rate)";
-    case MAKELONG(0x10de, 0x249d): return "NVIDIA GeForce RTX 3070 (mobile)";
-    case MAKELONG(0x10de, 0x2482): return "NVIDIA GeForce RTX 3070 Ti";
-    case MAKELONG(0x10de, 0x2206): return "NVIDIA GeForce RTX 3080 10GB";
-    case MAKELONG(0x10de, 0x2216): return "NVIDIA GeForce RTX 3080 10GB (Low Hash Rate)";
-    case MAKELONG(0x10de, 0x220a): return "NVIDIA GeForce RTX 3080 12GB";
-    case MAKELONG(0x10de, 0x2208): return "NVIDIA GeForce RTX 3080 Ti";
-    case MAKELONG(0x10de, 0x2204): return "NVIDIA GeForce RTX 3090";
-    case MAKELONG(0x10de, 0x2203): return "NVIDIA GeForce RTX 3090 Ti";
-    case MAKELONG(0x10de, 0x1eb8): return "NVIDIA Tesla T4";
-    case MAKELONG(0x10de, 0x2236): return "NVIDIA Ampere A10";
-    case MAKELONG(0x10de, 0x2882): return "NVIDIA GeForce RTX 4060";
-    case MAKELONG(0x10de, 0x28a0): return "NVIDIA GeForce RTX 4060M";
-    case MAKELONG(0x10de, 0x2803): return "NVIDIA GeForce RTX 4060 Ti 8GB";
-    case MAKELONG(0x10de, 0x2805): return "NVIDIA GeForce RTX 4060 Ti 16GB";
-    case MAKELONG(0x10de, 0x2786): return "NVIDIA GeForce RTX 4070";
-    case MAKELONG(0x10de, 0x2783): return "NVIDIA GeForce RTX 4070 SUPER";
-    case MAKELONG(0x10de, 0x2782): return "NVIDIA GeForce RTX 4070 Ti";
-    case MAKELONG(0x10de, 0x2705): return "NVIDIA GeForce RTX 4070 Ti SUPER";
-    case MAKELONG(0x10de, 0x2704): return "NVIDIA GeForce RTX 4080";
-    case MAKELONG(0x10de, 0x2702): return "NVIDIA GeForce RTX 4080 SUPER";
-    case MAKELONG(0x10de, 0x2684): return "NVIDIA GeForce RTX 4090";
-
-    /* AMD cards */
-    case MAKELONG(0x1002, 0x5246): return "ATI Rage Fury";
-    case MAKELONG(0x1002, 0x5144): return "ATI RADEON 7200 SERIES";
-    case MAKELONG(0x1002, 0x514c): return "ATI RADEON 8500 SERIES";
-    case MAKELONG(0x1002, 0x4144): return "ATI Radeon 9500";
-    case MAKELONG(0x1002, 0x5955): return "ATI RADEON XPRESS 200M Series";
-    case MAKELONG(0x1002, 0x5e4c): return "ATI Radeon X700 SE";
-    case MAKELONG(0x1002, 0x71c2): return "ATI Radeon X1600 Series";
-    case MAKELONG(0x1002, 0x94c7): return "ATI Mobility Radeon HD 2350";
-    case MAKELONG(0x1002, 0x9581): return "ATI Mobility Radeon HD 2600";
-    case MAKELONG(0x1002, 0x9400): return "ATI Radeon HD 2900 XT";
-    case MAKELONG(0x1002, 0x9620): return "ATI Radeon HD 3200 Graphics";
-    case MAKELONG(0x1002, 0x9515): return "ATI Radeon HD 3850 AGP";
-    case MAKELONG(0x1002, 0x9712): return "ATI Mobility Radeon HD 4200";
-    case MAKELONG(0x1002, 0x954f): return "ATI Radeon HD 4350";
-    case MAKELONG(0x1002, 0x9495): return "ATI Radeon HD 4600 Series";
-    case MAKELONG(0x1002, 0x944e): return "ATI Radeon HD 4700 Series";
-    case MAKELONG(0x1002, 0x944c): return "ATI Radeon HD 4800 Series";
-    case MAKELONG(0x1002, 0x68f9): return "ATI Radeon HD 5400 Series";
-    case MAKELONG(0x1002, 0x68d8): return "ATI Radeon HD 5600 Series";
-    case MAKELONG(0x1002, 0x68be): return "ATI Radeon HD 5700 Series";
-    case MAKELONG(0x1002, 0x6898): return "ATI Radeon HD 5800 Series";
-    case MAKELONG(0x1002, 0x689c): return "ATI Radeon HD 5900 Series";
-    case MAKELONG(0x1002, 0x9803): return "AMD Radeon HD 6300 series Graphics";
-    case MAKELONG(0x1002, 0x6770): return "AMD Radeon HD 6400 Series";
-    case MAKELONG(0x1002, 0x9644): return "AMD Radeon HD 6410D";
-    case MAKELONG(0x1002, 0x9648): return "AMD Radeon HD 6480G";
-    case MAKELONG(0x1002, 0x6760): return "AMD Radeon HD 6490M";
-    case MAKELONG(0x1002, 0x9640): return "AMD Radeon HD 6550D";
-    case MAKELONG(0x1002, 0x6758): return "AMD Radeon HD 6600 Series";
-    case MAKELONG(0x1002, 0x6741): return "AMD Radeon HD 6600M Series";
-    case MAKELONG(0x1002, 0x68ba): return "AMD Radeon HD 6700 Series";
-    case MAKELONG(0x1002, 0x6739): return "AMD Radeon HD 6800 Series";
-    case MAKELONG(0x1002, 0x6719): return "AMD Radeon HD 6900 Series";
-    case MAKELONG(0x1002, 0x9901): return "AMD Radeon HD 7660D";
-    case MAKELONG(0x1002, 0x683d): return "AMD Radeon HD 7700 Series";
-    case MAKELONG(0x1002, 0x6819): return "AMD Radeon HD 7800 Series";
-    case MAKELONG(0x1002, 0x6818): return "AMD Radeon HD 7870 Series";
-    case MAKELONG(0x1002, 0x679a): return "AMD Radeon HD 7900 Series";
-    case MAKELONG(0x1002, 0x6660): return "AMD Radeon HD 8600M Series";
-    case MAKELONG(0x1002, 0x6610): return "AMD Radeon HD 8670";
-    case MAKELONG(0x1002, 0x665c): return "AMD Radeon HD 8770";
-    case MAKELONG(0x1002, 0x9830): return "AMD Radeon HD 8400 / R3 Series";
-    case MAKELONG(0x1002, 0x130f): return "AMD Radeon(TM) R7 Graphics";
-    case MAKELONG(0x1002, 0x6939): return "AMD Radeon R9 285";
-    case MAKELONG(0x1002, 0x67b1): return "AMD Radeon R9 290";
-    case MAKELONG(0x1002, 0x67b0): return "AMD Radeon R9 290X";
-    case MAKELONG(0x1002, 0x7300): return "AMD Radeon (TM) R9 Fury Series";
-    case MAKELONG(0x1002, 0x6821): return "AMD Radeon R9 M370X";
-    case MAKELONG(0x1002, 0x6647): return "AMD Radeon R9 M380";
-    case MAKELONG(0x1002, 0x6920): return "AMD Radeon R9 M395X";
-    case MAKELONG(0x1002, 0x67ef): return "Radeon(TM) RX 460 Graphics";
-    case MAKELONG(0x1002, 0x67df): return "Radeon (TM) RX 480 Graphics";
-    case MAKELONG(0x1002, 0x73df): return "AMD Radeon RX 6700 XT";
-    case MAKELONG(0x1002, 0x687f): return "Radeon RX Vega";
-    case MAKELONG(0x1002, 0x69af): return "Radeon Pro Vega 20";
-    case MAKELONG(0x1002, 0x15dd): return "AMD Radeon(TM) Vega 10 Mobile Graphics";
-    case MAKELONG(0x1002, 0x66af): return "Radeon RX Vega 20";
-    case MAKELONG(0x1002, 0x731f): return "Radeon RX 5700 / 5700 XT";
-    case MAKELONG(0x1002, 0x7340): return "Radeon RX 5500M";
-    case MAKELONG(0x1002, 0x73bf): return "Radeon RX 6800/6800 XT / 6900 XT";
-    case MAKELONG(0x1002, 0x7480): return "AMD Radeon RX 7600 XT";
-    case MAKELONG(0x1002, 0x7590): return "AMD Radeon RX 9060 XT";
-    case MAKELONG(0x1002, 0x73a1): return "Radeon Pro V620";
-    case MAKELONG(0x1002, 0x73ae): return "Radeon Pro V620 VF";
-    case MAKELONG(0x1002, 0x163f): return "AMD VANGOGH";
-    case MAKELONG(0x1002, 0x164e): return "AMD Radeon(TM) Graphics";
-
-    /* Intel cards */
-    case MAKELONG(0x8086, 0x3577): return "Intel(R) 82830M Graphics Controller";
-    case MAKELONG(0x8086, 0x3582): return "Intel(R) 82852/82855 GM/GME Graphics Controller";
-    case MAKELONG(0x8086, 0x2562): return "Intel(R) 845G";
-    case MAKELONG(0x8086, 0x2572): return "Intel(R) 82865G Graphics Controller";
-    case MAKELONG(0x8086, 0x2582): return "Intel(R) 82915G/GV/910GL Express Chipset Family";
-    case MAKELONG(0x8086, 0x258a): return "Intel(R) E7221G";
-    case MAKELONG(0x8086, 0x2592): return "Mobile Intel(R) 915GM/GMS,910GML Express Chipset Family";
-    case MAKELONG(0x8086, 0x2772): return "Intel(R) 945G";
-    case MAKELONG(0x8086, 0x27a2): return "Mobile Intel(R) 945GM Express Chipset Family";
-    case MAKELONG(0x8086, 0x27ae): return "Intel(R) 945GME";
-    case MAKELONG(0x8086, 0x29b2): return "Intel(R) Q35";
-    case MAKELONG(0x8086, 0x29c2): return "Intel(R) G33";
-    case MAKELONG(0x8086, 0x29d2): return "Intel(R) Q33";
-    case MAKELONG(0x8086, 0xa001): return "Intel(R) IGD";
-    case MAKELONG(0x8086, 0xa011): return "Intel(R) IGD";
-    case MAKELONG(0x8086, 0x2992): return "Intel(R) 965Q";
-    case MAKELONG(0x8086, 0x2982): return "Intel(R) 965G";
-    case MAKELONG(0x8086, 0x2972): return "Intel(R) 946GZ";
-    case MAKELONG(0x8086, 0x2a02): return "Mobile Intel(R) 965 Express Chipset Family";
-    case MAKELONG(0x8086, 0x2a12): return "Intel(R) 965GME";
-    case MAKELONG(0x8086, 0x2a42): return "Mobile Intel(R) GM45 Express Chipset Family";
-    case MAKELONG(0x8086, 0x2e02): return "Intel(R) Integrated Graphics Device";
-    case MAKELONG(0x8086, 0x2e22): return "Intel(R) G45/G43";
-    case MAKELONG(0x8086, 0x2e12): return "Intel(R) Q45/Q43";
-    case MAKELONG(0x8086, 0x2e32): return "Intel(R) G41";
-    case MAKELONG(0x8086, 0x2e92): return "Intel(R) B43";
-    case MAKELONG(0x8086, 0x0042): return "Intel(R) HD Graphics";
-    case MAKELONG(0x8086, 0x0046): return "Intel(R) HD Graphics";
-    case MAKELONG(0x8086, 0x0122): return "Intel(R) HD Graphics 3000";
-    case MAKELONG(0x8086, 0x0126): return "Intel(R) HD Graphics 3000";
-    case MAKELONG(0x8086, 0x010a): return "Intel(R) HD Graphics Family";
-    case MAKELONG(0x8086, 0x0162): return "Intel(R) HD Graphics 4000";
-    case MAKELONG(0x8086, 0x0166): return "Intel(R) HD Graphics 4000";
-    case MAKELONG(0x8086, 0x015a): return "Intel(R) HD Graphics Family";
-    case MAKELONG(0x8086, 0x0412): return "Intel(R) HD Graphics 4600";
-    case MAKELONG(0x8086, 0x0416): return "Intel(R) HD Graphics 4600";
-    case MAKELONG(0x8086, 0x0a26): return "Intel(R) HD Graphics 5000";
-    case MAKELONG(0x8086, 0x0422): return "Intel(R) HD Graphics 5000";
-    case MAKELONG(0x8086, 0x0a22): return "Intel(R) Iris(TM) Graphics 5100";
-    case MAKELONG(0x8086, 0x0a2a): return "Intel(R) Iris(TM) Graphics 5100";
-    case MAKELONG(0x8086, 0x0a2b): return "Intel(R) Iris(TM) Graphics 5100";
-    case MAKELONG(0x8086, 0x0a2e): return "Intel(R) Iris(TM) Graphics 5100";
-    case MAKELONG(0x8086, 0x0d22): return "Intel(R) Iris(TM) Pro Graphics 5200";
-    case MAKELONG(0x8086, 0x0d26): return "Intel(R) Iris(TM) Pro Graphics 5200";
-    case MAKELONG(0x8086, 0x0d2a): return "Intel(R) Iris(TM) Pro Graphics 5200";
-    case MAKELONG(0x8086, 0x0d2b): return "Intel(R) Iris(TM) Pro Graphics 5200";
-    case MAKELONG(0x8086, 0x0d2e): return "Intel(R) Iris(TM) Pro Graphics 5200";
-    case MAKELONG(0x8086, 0x0c22): return "Intel(R) Iris(TM) Pro Graphics 5200";
-    case MAKELONG(0x8086, 0x161e): return "Intel(R) HD Graphics 5300";
-    case MAKELONG(0x8086, 0x1616): return "Intel(R) HD Graphics 5500";
-    case MAKELONG(0x8086, 0x1612): return "Intel(R) HD Graphics 5600";
-    case MAKELONG(0x8086, 0x1626): return "Intel(R) HD Graphics 6000";
-    case MAKELONG(0x8086, 0x162b): return "Intel(R) Iris(TM) Graphics 6100";
-    case MAKELONG(0x8086, 0x1622): return "Intel(R) Iris(TM) Pro Graphics 6200";
-    case MAKELONG(0x8086, 0x162a): return "Intel(R) Iris(TM) Pro Graphics P6300";
-    case MAKELONG(0x8086, 0x1902): return "Intel(R) HD Graphics 510";
-    case MAKELONG(0x8086, 0x1906): return "Intel(R) HD Graphics 510";
-    case MAKELONG(0x8086, 0x190b): return "Intel(R) HD Graphics 510";
-    case MAKELONG(0x8086, 0x191e): return "Intel(R) HD Graphics 515";
-    case MAKELONG(0x8086, 0x1916): return "Intel(R) HD Graphics 520";
-    case MAKELONG(0x8086, 0x1921): return "Intel(R) HD Graphics 520";
-    case MAKELONG(0x8086, 0x1912): return "Intel(R) HD Graphics 530";
-    case MAKELONG(0x8086, 0x191b): return "Intel(R) HD Graphics 530";
-    case MAKELONG(0x8086, 0x191d): return "Intel(R) HD Graphics P530";
-    case MAKELONG(0x8086, 0x1926): return "Intel(R) Iris(TM) Graphics 540";
-    case MAKELONG(0x8086, 0x1927): return "Intel(R) Iris(TM) Graphics 550";
-    case MAKELONG(0x8086, 0x192b): return "Intel(R) Iris(TM) Graphics 555";
-    case MAKELONG(0x8086, 0x192d): return "Intel(R) Iris(TM) Graphics P555";
-    case MAKELONG(0x8086, 0x1932): return "Intel(R) Iris(TM) Pro Graphics 580";
-    case MAKELONG(0x8086, 0x193b): return "Intel(R) Iris(TM) Pro Graphics 580";
-    case MAKELONG(0x8086, 0x193a): return "Intel(R) Iris(TM) Pro Graphics P580";
-    case MAKELONG(0x8086, 0x193d): return "Intel(R) Iris(TM) Pro Graphics P580";
-    case MAKELONG(0x8086, 0x87c0): return "Intel(R) UHD Graphics 617";
-    case MAKELONG(0x8086, 0x3ea0): return "Intel(R) UHD Graphics 620";
-    case MAKELONG(0x8086, 0x591e): return "Intel(R) HD Graphics 615";
-    case MAKELONG(0x8086, 0x5916): return "Intel(R) HD Graphics 620";
-    case MAKELONG(0x8086, 0x5912): return "Intel(R) HD Graphics 630";
-    case MAKELONG(0x8086, 0x591b): return "Intel(R) HD Graphics 630";
-    case MAKELONG(0x8086, 0x3e9b): return "Intel(R) UHD Graphics 630";
-    case MAKELONG(0x8086, 0x3e91): return "Intel(R) UHD Graphics 630";
-    }
-
-    if (!default_name) return "Wine Adapter";
-    return default_name;
-};
-
 static BOOL write_gpu_to_registry( const struct gpu *gpu, const struct pci_id *pci,
                                    ULONGLONG memory_size )
 {
-    unsigned int size, name_size = (wcslen( gpu->name ) + 1) * sizeof(WCHAR);
+    const WCHAR *desc;
     char buffer[4096], *tmp;
     WCHAR bufferW[512];
+    unsigned int size;
     HKEY subkey;
     LARGE_INTEGER ft;
     ULONG value;
     HKEY hkey;
 
     static const BOOL present = TRUE;
+    static const WCHAR wine_adapterW[] = {'W','i','n','e',' ','A','d','a','p','t','e','r',0};
     static const WCHAR driver_date_dataW[] =
         {'D','r','i','v','e','r','D','a','t','e','D','a','t','a',0};
     static const WCHAR adapter_stringW[] =
@@ -1545,7 +1192,7 @@
 
     if ((subkey = reg_create_ascii_key( hkey, devpkey_device_driver_desc, 0, NULL )))
     {
-        set_reg_value( subkey, NULL, 0xffff0000 | DEVPROP_TYPE_STRING, gpu->name, name_size );
+        set_reg_value( subkey, NULL, 0xffff0000 | DEVPROP_TYPE_STRING, gpu->name, (wcslen( gpu->name ) + 1) * sizeof(WCHAR) );
         NtClose( subkey );
     }
 
@@ -1575,7 +1222,9 @@
         NtClose( subkey );
     }
 
-    set_reg_value( hkey, device_descW, REG_SZ, gpu->name, name_size );
+    desc = gpu->name;
+    if (!desc[0]) desc = wine_adapterW;
+    set_reg_value( hkey, device_descW, REG_SZ, desc, (lstrlenW( desc ) + 1) * sizeof(WCHAR) );
 
     if ((subkey = reg_create_ascii_key( hkey, "Device Parameters", 0, NULL )))
     {
@@ -1611,11 +1260,14 @@
     if (!(hkey = reg_create_ascii_key( control_key, buffer, 0, NULL ))) return FALSE;
 
     set_reg_value( hkey, driver_dateW, REG_SZ, bufferW, format_date( bufferW, ft.QuadPart ));
+
     set_reg_value( hkey, driver_date_dataW, REG_BINARY, &ft, sizeof(ft) );
-    set_reg_value( hkey, driver_descW, REG_SZ, gpu->name, name_size );
-    set_reg_value( hkey, adapter_stringW, REG_SZ, gpu->name, name_size );
-    set_reg_value( hkey, bios_stringW, REG_SZ, gpu->name, name_size );
-    set_reg_value( hkey, chip_typeW, REG_SZ, gpu->name, name_size );
+
+    size = (lstrlenW( desc ) + 1) * sizeof(WCHAR);
+    set_reg_value( hkey, driver_descW, REG_SZ, desc, size );
+    set_reg_value( hkey, adapter_stringW, REG_SZ, desc, size );
+    set_reg_value( hkey, bios_stringW, REG_SZ, desc, size );
+    set_reg_value( hkey, chip_typeW, REG_SZ, desc, size );
     set_reg_value( hkey, dac_typeW, REG_SZ, ramdacW, sizeof(ramdacW) );
 
     /* If we failed to retrieve the gpu memory size set a default of 1Gb */
@@ -1626,6 +1278,7 @@
     set_reg_value( hkey, memory_sizeW, REG_DWORD, &value, sizeof(value) );
 
     set_reg_ascii_value( hkey, "DriverVersion", driver_vendor_to_version( pci->vendor ) );
+
     NtClose( hkey );
 
 
@@ -1643,7 +1296,7 @@
         asciiz_to_unicode( bufferW, "DriverVersion" );
         set_reg_value( hkey, bufferW, REG_QWORD, &ver, sizeof(ver) );
         asciiz_to_unicode( bufferW, "Description" );
-        set_reg_value( hkey, bufferW, REG_SZ, gpu->name, name_size );
+        set_reg_value( hkey, bufferW, REG_SZ, desc, (lstrlenW( desc ) + 1) * sizeof(WCHAR) );
         if (pci->vendor && pci->device)
         {
             asciiz_to_unicode( bufferW, "DeviceId" );
@@ -1678,6 +1331,38 @@
     return NULL;
 }
 
+static void fixup_device_id( const struct pci_id **pci_id )
+{
+    static struct pci_id fake_id;
+    const char *sgi;
+
+    if ((*pci_id)->vendor == 0x10de /* NVIDIA */ && (sgi = getenv("WINE_HIDE_NVIDIA_GPU")) && *sgi != '0')
+    {
+        fake_id.vendor = 0x1002; /* AMD */
+        fake_id.device = 0x73df; /* RX 6700XT */
+        *pci_id = &fake_id;
+    }
+    else if ((*pci_id)->vendor == 0x1002 /* AMD */ && (sgi = getenv("WINE_HIDE_AMD_GPU")) && *sgi != '0')
+    {
+        fake_id.vendor = 0x10de; /* NVIDIA */
+        fake_id.device = 0x2487; /* RTX 3060 */
+        *pci_id = &fake_id;
+    }
+    else if ((*pci_id)->vendor == 0x1002 && ((*pci_id)->device == 0x163f || (*pci_id)->device == 0x1435)
+             && (sgi = getenv("WINE_HIDE_VANGOGH_GPU")) && *sgi != '0')
+    {
+        fake_id.vendor = (*pci_id)->vendor;
+        fake_id.device = 0x687f; /* Radeon RX Vega 56/64 */
+        *pci_id = &fake_id;
+    }
+    else if ((*pci_id)->vendor == 0x8086 /* Intel */ && (sgi = getenv("WINE_HIDE_INTEL_GPU")) && *sgi != '0')
+    {
+        fake_id.vendor = 0x1002; /* AMD */
+        fake_id.device = 0x73df; /* RX 6700XT */
+        *pci_id = &fake_id;
+    }
+}
+
 static void add_gpu( const char *name, const struct pci_id *pci_id, const GUID *vulkan_uuid, void *param )
 {
     struct device_manager_ctx *ctx = param;
@@ -1727,10 +1412,10 @@
     else if (vulkan_gpu) gpu->vulkan_uuid = vulkan_gpu->uuid;
 
     if (!pci_id->vendor && !pci_id->device && vulkan_gpu) pci_id = &vulkan_gpu->pci_id;
+    fixup_device_id( &pci_id );
 
-    name = gpu_device_name( pci_id->vendor, pci_id->device, name );
-    if (!strcmp( name, "Wine Adapter" ) && vulkan_gpu) name = vulkan_gpu->name;
-    RtlUTF8ToUnicodeN( gpu->name, sizeof(gpu->name) - sizeof(WCHAR), &len, name, strlen( name ) );
+    if ((!name || !strcmp( name, "Wine GPU" )) && vulkan_gpu) name = vulkan_gpu->name;
+    if (name) RtlUTF8ToUnicodeN( gpu->name, sizeof(gpu->name) - sizeof(WCHAR), &len, name, strlen( name ) );
 
     snprintf( gpu->path, sizeof(gpu->path), "PCI\\VEN_%04X&DEV_%04X&SUBSYS_%08X&REV_%02X\\%08X",
               pci_id->vendor, pci_id->device, pci_id->subsystem, pci_id->revision, gpu->index );
@@ -1761,12 +1446,12 @@
         if (query_reg_value( subkey, NULL, value, sizeof(buffer) ) != sizeof(LUID))
         {
             NtAllocateLocallyUniqueId( &gpu->luid );
-            TRACE("allocated luid %08x%08x\n", gpu->luid.HighPart, gpu->luid.LowPart );
+            TRACE("allocated luid %08x%08x\n", (int)gpu->luid.HighPart, (int)gpu->luid.LowPart );
         }
         else
         {
             memcpy( &gpu->luid, value->Data, sizeof(gpu->luid) );
-            TRACE("got luid %08x%08x\n", gpu->luid.HighPart, gpu->luid.LowPart );
+            TRACE("got luid %08x%08x\n", (int)gpu->luid.HighPart, (int)gpu->luid.LowPart );
         }
         NtClose( subkey );
     }
@@ -1940,14 +1625,6 @@
         NtClose( subkey );
     }
 
-    /* WINE_DEVPROPKEY_MONITOR_HDR_ENABLED */
-    if ((subkey = reg_create_ascii_key( hkey, wine_devpropkey_monitor_hdr_enabledA, 0, NULL )))
-    {
-        set_reg_value( subkey, NULL, 0xffff0000 | DEVPROP_TYPE_BOOLEAN,
-                       &monitor->hdr_enabled, sizeof(monitor->hdr_enabled) );
-        NtClose( subkey );
-    }
-
     NtClose( hkey );
 
 
@@ -1977,7 +1654,6 @@
     monitor->id = source->monitor_count;
     monitor->output_id = ctx->monitor_count;
     monitor->rc_work = gdi_monitor->rc_work;
-    monitor->hdr_enabled = gdi_monitor->hdr_enabled;
 
     TRACE( "%u %s %s\n", monitor->id, wine_dbgstr_rect(&gdi_monitor->rc_monitor), wine_dbgstr_rect(&gdi_monitor->rc_work) );
 
@@ -2010,34 +1686,27 @@
     UINT i = 0;
 
     while (i < count && memcmp( sizes + i, &size, sizeof(size) )) i++;
-    if (i < count) return 0;
-
-    TRACE( "adding size %s\n", wine_dbgstr_point((POINT *)&size) );
-    sizes[i] = size;
-    return 1;
-}
-
-static UINT add_virtual_mode( DEVMODEW *modes, UINT count, const DEVMODEW *mode )
-{
-    TRACE( "adding mode %s\n", debugstr_devmodew(mode) );
-    modes[count] = *mode;
-    return 1;
+    if (i == count) sizes[i] = size;
+    return i == count ? 1 : 0;
 }
 
 static SIZE *get_screen_sizes( const DEVMODEW *maximum, const DEVMODEW *modes, UINT modes_count,
                                UINT *sizes_count )
 {
-    static SIZE default_sizes[] =
+    static SIZE lowres_sizes[] =
     {
         /* 4:3 */
         { 640,  480},
+        /* 16:9 */
+        { 960,  540},
+    };
+    static SIZE default_sizes[] =
+    {
+        /* 4:3 */
         { 800,  600},
         {1024,  768},
-        {1152,  864},
-        {1280,  960},
         {1600, 1200},
         /* 16:9 */
-        { 960,  540},
         {1280,  720},
         {1600,  900},
         {1920, 1080},
@@ -2059,15 +1728,16 @@
         {3840, 1600},
         /* 5:4 */
         {1280, 1024},
-        /* 5:3 */
-        {1280,  768},
     };
     UINT max_width = devmode_get( maximum, DM_PELSWIDTH ), max_height = devmode_get( maximum, DM_PELSHEIGHT );
     SIZE *sizes, max_size = {.cx = max( max_width, max_height ), .cy = min( max_width, max_height )};
     const DEVMODEW *mode;
+    BOOL enable_lowres;
     UINT i, count;
 
-    count = 1 + ARRAY_SIZE(default_sizes) + modes_count;
+    const char *env;
+
+    count = 1 + ARRAY_SIZE(default_sizes) + ARRAY_SIZE(lowres_sizes) + modes_count;
     if (!(sizes = malloc( count * sizeof(*sizes) ))) return NULL;
 
     count = add_screen_size( sizes, 0, max_size );
@@ -2077,19 +1747,39 @@
         count += add_screen_size( sizes, count, default_sizes[i] );
     }
 
+    /* Titan Souls renders incorrectly if we report modes smaller than 800x600 */
+    if ((enable_lowres = (!(env = getenv( "SteamAppId" )) || (strcmp( env, "297130" ) && strcmp( env, "403640" )))))
+    {
+        memcpy( sizes + count, lowres_sizes, ARRAY_SIZE(lowres_sizes) * sizeof(*sizes) );
+        count += ARRAY_SIZE(lowres_sizes);
+    }
+
     for (mode = modes; mode && modes_count; mode = NEXT_DEVMODEW(mode), modes_count--)
     {
         UINT width = devmode_get( mode, DM_PELSWIDTH ), height = devmode_get( mode, DM_PELSHEIGHT );
         SIZE size = {.cx = max( width, height ), .cy = min( width, height )};
-        if (!size.cx || size.cx > max_size.cx) continue;
-        if (!size.cy || size.cy > max_size.cy) continue;
+        if (!size.cx || (size.cx < 800 && !enable_lowres) || size.cx > max_size.cx) continue;
+        if (!size.cy || (size.cy < 600 && !enable_lowres) || size.cy > max_size.cy) continue;
         count += add_screen_size( sizes, count, size );
     }
 
+    if (limit_resolutions && count > limit_resolutions) count = limit_resolutions;
     *sizes_count = count;
     return sizes;
 }
 
+static UINT add_virtual_mode( DEVMODEW *modes, UINT count, const DEVMODEW *mode, BOOL center )
+{
+    TRACE( "adding %s\n", debugstr_devmodew(mode) );
+    modes[count++] = *mode;
+    if (!center) return 1;
+
+    modes[count] = *mode;
+    modes[count].dmFields |= DM_DISPLAYFIXEDOUTPUT;
+    modes[count].dmDisplayFixedOutput = DMDFO_CENTER;
+    return 2;
+}
+
 static DEVMODEW *get_virtual_modes( const DEVMODEW *initial, const DEVMODEW *maximum,
                                     const DEVMODEW *host_modes, UINT host_modes_count, UINT32 *modes_count )
 {
@@ -2098,16 +1788,39 @@
     SIZE *screen_sizes;
     BOOL vertical;
 
+    BOOL center_modes = FALSE;
+    const char *env;
+
     /* Check the ratio of dmPelsWidth to dmPelsHeight to determine whether the initial display mode
      * is in horizontal or vertical orientation. DMDO_DEFAULT is the natural orientation of the
      * device, which isn't necessarily a horizontal mode */
     vertical = initial->dmPelsHeight > initial->dmPelsWidth;
 
+    if ((env = getenv( "WINE_CENTER_DISPLAY_MODES" )))
+        center_modes = (env[0] != '0');
+    else if ((env = getenv( "SteamAppId" )))
+        center_modes = !strcmp( env, "359870" );
+
     freqs[1] = devmode_get( initial, DM_DISPLAYFREQUENCY );
     if (freqs[1] <= 60) freqs[1] = 0;
 
     if (!(screen_sizes = get_screen_sizes( maximum, host_modes, host_modes_count, &sizes_count ))) return NULL;
-    modes = malloc( ARRAY_SIZE(freqs) * ARRAY_SIZE(depths) * (sizes_count + 2) * sizeof(*modes) );
+    modes = malloc( (2 * ARRAY_SIZE(freqs) * ARRAY_SIZE(depths) * (sizes_count + 2) + 1) * sizeof(*modes) );
+
+    if ((env = getenv( "SteamAppId" )) && !strcmp( env, "403640" ))
+    {
+        DEVMODEW mode =
+        {
+            .dmSize = sizeof(mode),
+            .dmFields = DM_DISPLAYORIENTATION | DM_BITSPERPEL | DM_PELSWIDTH | DM_PELSHEIGHT | DM_DISPLAYFLAGS | DM_DISPLAYFREQUENCY,
+            .dmDisplayFrequency = 30,
+            .dmBitsPerPel = 32,
+            .dmDisplayOrientation = initial->dmDisplayOrientation,
+            .dmPelsWidth = 800,
+            .dmPelsHeight = 600,
+        };
+        count += add_virtual_mode( modes, count, &mode, center_modes );
+    }
 
     for (i = 0; modes && i < ARRAY_SIZE(depths); ++i)
     for (f = 0; f < ARRAY_SIZE(freqs); ++f)
@@ -2130,23 +1843,24 @@
             if (mode.dmPelsWidth > maximum->dmPelsWidth || mode.dmPelsHeight > maximum->dmPelsHeight) continue;
             if (mode.dmPelsWidth == maximum->dmPelsWidth && mode.dmPelsHeight == maximum->dmPelsHeight) continue;
             if (mode.dmPelsWidth == initial->dmPelsWidth && mode.dmPelsHeight == initial->dmPelsHeight) continue;
-            count += add_virtual_mode( modes, count, &mode );
+            count += add_virtual_mode( modes, count, &mode, center_modes );
         }
 
         mode.dmPelsWidth = vertical ? initial->dmPelsHeight : initial->dmPelsWidth;
         mode.dmPelsHeight = vertical ? initial->dmPelsWidth : initial->dmPelsHeight;
-        count += add_virtual_mode( modes, count, &mode );
+        count += add_virtual_mode( modes, count, &mode, center_modes );
 
         if (maximum->dmPelsWidth != initial->dmPelsWidth || maximum->dmPelsHeight != initial->dmPelsHeight)
         {
             mode.dmPelsWidth = vertical ? maximum->dmPelsHeight : maximum->dmPelsWidth;
             mode.dmPelsHeight = vertical ? maximum->dmPelsWidth : maximum->dmPelsHeight;
-            count += add_virtual_mode( modes, count, &mode );
+            count += add_virtual_mode( modes, count, &mode, center_modes );
         }
     }
 
-    *modes_count = count;
     free( screen_sizes );
+
+    *modes_count = count;
     return modes;
 }
 
@@ -2169,6 +1883,24 @@
         modes = current;
         modes_count = 1;
     }
+    else if (emulate_modelist)
+    {
+        physical = *current;
+        if ((virtual_modes = get_virtual_modes( &physical, &physical, host_modes, host_modes_count, &virtual_count )))
+        {
+            modes_count = virtual_count;
+            modes = virtual_modes;
+        }
+
+        /* HACK: Gamescope doesn't really changes the display mode, pretend it changed to what was requested */
+        if (user_driver->pHasWindowManager( "steamcompmgr" ) && read_source_mode( source->key, ENUM_CURRENT_SETTINGS, &virtual )
+            && !is_detached_mode( &virtual ))
+        {
+            WARN( "Faking current mode to %s\n", debugstr_devmodew(&virtual) );
+            current = &virtual;
+            detached = *current;
+        }
+    }
 
     physical = modes_count == 1 ? *modes : *current;
     if (ctx->is_primary) ctx->primary = *current;
@@ -2180,7 +1912,7 @@
     if (modes_count > 1 || current == &detached)
     {
         reg_delete_value( source->key, physicalW );
-        virtual_modes = NULL;
+        if (!emulate_modelist) virtual_modes = NULL;
     }
     else
     {
@@ -2257,11 +1989,27 @@
     return !!(source->state_flags & DISPLAY_DEVICE_PRIMARY_DEVICE);
 }
 
+static UINT gcd( UINT a, UINT b )
+{
+    int r;
+
+    while (1)
+    {
+        if (!a) return b;
+        if (!b) return a;
+        r = a % b;
+        a = b;
+        b = r;
+    }
+}
+
 /* display_lock must be held */
 static void monitor_virt_to_raw_ratio( struct monitor *monitor, UINT *num, UINT *den )
 {
     struct source *source = monitor->source;
 
+    if (!num || !den) return;
+
     *num = *den = 1;
     if (!source) return;
 
@@ -2307,7 +2055,7 @@
 
     if (monitor->source)
     {
-        double points[4] = {rect.left, rect.top, rect.right, rect.bottom}, from[2], to[2];
+        float points[4] = {rect.left, rect.top, rect.right, rect.bottom}, from[2], to[2];
         DEVMODEW current_mode = {.dmSize = sizeof(DEVMODEW)}, physical_mode;
         UINT num, den, dpi;
 
@@ -2337,16 +2085,14 @@
 
         for (int i = 0; i < ARRAY_SIZE(points); i++)
         {
-            points[i] *= (double)dpi / dpi_from;
+            points[i] *= (float)dpi / dpi_from;
             points[i] -= from[i & 1];
-            points[i] *= (double)num / den;
+            points[i] *= (float)num / den;
             points[i] += to[i & 1];
-            points[i] *= (double)dpi_to / dpi;
-            points[i] = roundf( points[i] );
-            points[i] = min( INT_MAX, max( INT_MIN, (INT64)points[i] ));
+            points[i] *= (float)dpi_to / dpi;
         }
 
-        SetRect( &rect, points[0], points[1], points[2], points[3] );
+        SetRect( &rect, round( points[0] ), round( points[1] ), round( points[2] ), round( points[3] ) );
         return rect;
     }
 
@@ -2527,7 +2273,7 @@
     HKEY hkey;
     BOOL ret;
 
-    /* If user driver didn't initialize the registry, then exit */
+    /* If user driver did initialize the registry, then exit */
     if (!enum_key && !(enum_key = reg_open_ascii_key( NULL, enum_keyA )))
         return FALSE;
     if (!video_key && !(video_key = reg_open_ascii_key( NULL, devicemap_video_keyA )))
@@ -2636,7 +2382,7 @@
     DEVMODEW mode = {.dmSize = sizeof(mode)};
     struct source *source;
 
-    add_gpu( NULL, &pci_id, NULL, ctx );
+    add_gpu( "Wine GPU", &pci_id, NULL, ctx );
     add_source( "Default", source_flags, system_dpi, ctx );
 
     assert( !list_empty( &sources ) );
@@ -2818,6 +2564,8 @@
 
     init_display_driver(); /* make sure to load the driver before anything else */
 
+    if (user_driver->pHasWindowManager( "steamcompmgr" )) emulate_modeset = FALSE;
+
     pthread_mutex_lock( &display_lock );
 
     serial = get_monitor_update_serial();
@@ -2923,14 +2671,17 @@
     LIST_FOR_EACH_ENTRY(monitor, &monitors, struct monitor, entry)
     {
         RECT intersect, monitor_rect;
+        UINT density, raw_dpi, x, y;
 
         if (!is_monitor_active( monitor ) || monitor->is_clone) continue;
+        raw_dpi = monitor_get_dpi( monitor, MDT_RAW_DPI, &x, &y );
+        density = raw_dpi * raw_dpi / 96 / 96;
 
         monitor_rect = monitor_get_rect( monitor, dpi, type );
         if (intersect_rect( &intersect, &monitor_rect, &rect ))
         {
             /* check for larger intersecting area */
-            UINT area = (intersect.right - intersect.left) * (intersect.bottom - intersect.top);
+            UINT area = (intersect.right - intersect.left) * (intersect.bottom - intersect.top) * density;
             if (area > max_area)
             {
                 max_area = area;
@@ -3028,6 +2779,22 @@
     return rect;
 }
 
+RECT map_rect_virt_to_raw_for_monitor( HMONITOR handle, RECT rect, UINT dpi_from )
+{
+    struct monitor *monitor;
+    UINT num, den, x, y;
+
+    if (!lock_display_devices( FALSE )) return rect;
+    if ((monitor = get_monitor_from_handle( handle )))
+    {
+        monitor_virt_to_raw_ratio( monitor, &num, &den );
+        rect = map_dpi_rect( rect, dpi_from * den, monitor_get_dpi( monitor, MDT_DEFAULT, &x, &y ) * num );
+    }
+    unlock_display_devices();
+
+    return rect;
+}
+
 /* map (absolute) window rects from MDT_DEFAULT to MDT_RAW_DPI coordinates */
 struct window_rects map_window_rects_virt_to_raw( struct window_rects rects, UINT dpi_from )
 {
@@ -3075,7 +2842,7 @@
 /**********************************************************************
  *              get_win_monitor_dpi
  */
-UINT get_win_monitor_dpi( HWND hwnd, UINT *raw_dpi )
+UINT get_win_monitor_dpi( HWND hwnd, UINT *raw_dpi_num, UINT *raw_dpi_den )
 {
     UINT dpi = NTUSER_DPI_CONTEXT_GET_DPI( get_window_dpi_awareness_context( hwnd ) );
     HWND parent = get_parent( hwnd );
@@ -3088,7 +2855,7 @@
         return 0;
     }
 
-    if (win == WND_DESKTOP) return monitor_dpi_from_rect( rect, get_thread_dpi(), raw_dpi );
+    if (win == WND_DESKTOP) return monitor_dpi_from_rect( rect, get_thread_dpi(), raw_dpi_num, raw_dpi_den );
     if (win == WND_OTHER_PROCESS)
     {
         if (!get_window_rect( hwnd, &rect, dpi )) return 0;
@@ -3097,7 +2864,7 @@
     else if ((parent = win->parent) && parent != get_desktop_window())
     {
         release_win_ptr( win );
-        return get_win_monitor_dpi( parent, raw_dpi );
+        return get_win_monitor_dpi( parent, raw_dpi_num, raw_dpi_den );
     }
     else
     {
@@ -3105,7 +2872,7 @@
         release_win_ptr( win );
     }
 
-    return monitor_dpi_from_rect( rect, dpi, raw_dpi );
+    return monitor_dpi_from_rect( rect, dpi, raw_dpi_num, raw_dpi_den );
 }
 
 /* keep in sync with user32 */
@@ -3258,7 +3025,7 @@
  */
 static POINT point_win_to_phys_dpi( HWND hwnd, POINT pt )
 {
-    UINT raw_dpi, dpi = get_win_monitor_dpi( hwnd, &raw_dpi );
+    UINT dpi = get_win_monitor_dpi( hwnd, NULL, NULL );
     return map_dpi_point( pt, get_dpi_for_window( hwnd ), dpi );
 }
 
@@ -3267,7 +3034,7 @@
  */
 POINT point_phys_to_win_dpi( HWND hwnd, POINT pt )
 {
-    UINT raw_dpi, dpi = get_win_monitor_dpi( hwnd, &raw_dpi );
+    UINT dpi = get_win_monitor_dpi( hwnd, NULL, NULL );
     return map_dpi_point( pt, dpi, get_dpi_for_window( hwnd ) );
 }
 
@@ -3276,8 +3043,8 @@
  */
 POINT point_thread_to_win_dpi( HWND hwnd, POINT pt )
 {
-    UINT dpi = get_thread_dpi(), raw_dpi;
-    if (!dpi) dpi = get_win_monitor_dpi( hwnd, &raw_dpi );
+    UINT dpi = get_thread_dpi();
+    if (!dpi) dpi = get_win_monitor_dpi( hwnd, NULL, NULL );
     return map_dpi_point( pt, dpi, get_dpi_for_window( hwnd ));
 }
 
@@ -3286,8 +3053,8 @@
  */
 RECT rect_thread_to_win_dpi( HWND hwnd, RECT rect )
 {
-    UINT dpi = get_thread_dpi(), raw_dpi;
-    if (!dpi) dpi = get_win_monitor_dpi( hwnd, &raw_dpi );
+    UINT dpi = get_thread_dpi();
+    if (!dpi) dpi = get_win_monitor_dpi( hwnd, NULL, NULL );
     return map_dpi_rect( rect, dpi, get_dpi_for_window( hwnd ) );
 }
 
@@ -3315,6 +3082,32 @@
     return rect;
 }
 
+BOOL is_window_rect_full_screen( const RECT *rect, UINT dpi )
+{
+    struct monitor *monitor;
+    BOOL ret = FALSE;
+
+    if (!lock_display_devices( FALSE )) return FALSE;
+
+    LIST_FOR_EACH_ENTRY( monitor, &monitors, struct monitor, entry )
+    {
+        RECT monrect;
+
+        if (!is_monitor_active( monitor ) || monitor->is_clone) continue;
+
+        monrect = monitor_get_rect( monitor, dpi, MDT_DEFAULT );
+        if (rect->left <= monrect.left && rect->right >= monrect.right &&
+            rect->top <= monrect.top && rect->bottom >= monrect.bottom)
+        {
+            ret = TRUE;
+            break;
+        }
+    }
+
+    unlock_display_devices();
+    return ret;
+}
+
 static UINT get_display_index( const UNICODE_STRING *name )
 {
     static const WCHAR displayW[] = {'\\','\\','.','\\','D','I','S','P','L','A','Y'};
@@ -3819,7 +3612,7 @@
     struct source *source = NULL;
     BOOL found = FALSE;
 
-    TRACE( "%s %u %p %#x\n", debugstr_us( device ), index, info, flags );
+    TRACE( "%s %u %p %#x\n", debugstr_us( device ), (int)index, info, (int)flags );
 
     if (!info || !info->cb) return STATUS_UNSUCCESSFUL;
 
@@ -3945,19 +3738,19 @@
 {
     TRACE( "dmFields=%s ", _DM_fields(devmode->dmFields) );
     if (devmode->dmFields & DM_BITSPERPEL)
-        TRACE( "dmBitsPerPel=%u ", devmode->dmBitsPerPel );
+        TRACE( "dmBitsPerPel=%u ", (int)devmode->dmBitsPerPel );
     if (devmode->dmFields & DM_PELSWIDTH)
-        TRACE( "dmPelsWidth=%u ", devmode->dmPelsWidth );
+        TRACE( "dmPelsWidth=%u ", (int)devmode->dmPelsWidth );
     if (devmode->dmFields & DM_PELSHEIGHT)
-        TRACE( "dmPelsHeight=%u ", devmode->dmPelsHeight );
+        TRACE( "dmPelsHeight=%u ", (int)devmode->dmPelsHeight );
     if (devmode->dmFields & DM_DISPLAYFREQUENCY)
-        TRACE( "dmDisplayFrequency=%u ", devmode->dmDisplayFrequency );
+        TRACE( "dmDisplayFrequency=%u ", (int)devmode->dmDisplayFrequency );
     if (devmode->dmFields & DM_POSITION)
-        TRACE( "dmPosition=(%d,%d) ", devmode->dmPosition.x, devmode->dmPosition.y );
+        TRACE( "dmPosition=(%d,%d) ", (int)devmode->dmPosition.x, (int)devmode->dmPosition.y );
     if (devmode->dmFields & DM_DISPLAYFLAGS)
-        TRACE( "dmDisplayFlags=%#x ", devmode->dmDisplayFlags );
+        TRACE( "dmDisplayFlags=%#x ", (int)devmode->dmDisplayFlags );
     if (devmode->dmFields & DM_DISPLAYORIENTATION)
-        TRACE( "dmDisplayOrientation=%u ", devmode->dmDisplayOrientation );
+        TRACE( "dmDisplayOrientation=%u ", (int)devmode->dmDisplayOrientation );
     TRACE("\n");
 }
 
@@ -4021,7 +3814,7 @@
 
     if ((full_mode->dmFields & (DM_PELSWIDTH | DM_PELSHEIGHT)) != (DM_PELSWIDTH | DM_PELSHEIGHT))
     {
-        WARN( "devmode doesn't specify the resolution: %#x\n", full_mode->dmFields );
+        WARN( "devmode doesn't specify the resolution: %#x\n", (int)full_mode->dmFields );
         return FALSE;
     }
 
@@ -4339,7 +4132,7 @@
     struct source *primary, *source;
     DEVMODEW *mode, *displays;
     HWND restorer_window;
-    LONG ret;
+    UINT ret;
 
     if (!lock_display_devices( FALSE )) return DISP_CHANGE_FAILED;
     if (!(displays = get_display_settings( target, devmode )))
@@ -4368,7 +4161,10 @@
 
     /* use the default implementation in virtual desktop mode */
     if (is_virtual_desktop() || emulate_modeset) ret = DISP_CHANGE_SUCCESSFUL;
+    /* HACK: Gamescope doesn't really changes the display mode, no point even trying as it sometimes fails with emulated mode list */
+    else if (user_driver->pHasWindowManager( "steamcompmgr" )) ret = DISP_CHANGE_SUCCESSFUL;
     else ret = user_driver->pChangeDisplaySettings( displays, primary_name, hwnd, flags, lparam );
+    if (ret != DISP_CHANGE_SUCCESSFUL) WARN( "Failed to change display settings, ret %d\n", ret );
 
     if (ret == DISP_CHANGE_SUCCESSFUL)
     {
@@ -4408,7 +4204,7 @@
     int ret = DISP_CHANGE_SUCCESSFUL;
     struct source *source;
 
-    TRACE( "%s %p %p %#x %p\n", debugstr_us(devname), devmode, hwnd, flags, lparam );
+    TRACE( "%s %p %p %#x %p\n", debugstr_us(devname), devmode, hwnd, (int)flags, lparam );
     TRACE( "flags=%s\n", _CDS_flags(flags) );
 
     if ((!devname || !devname->Length) && !devmode) return apply_display_settings( NULL, NULL, hwnd, flags, lparam );
@@ -4468,7 +4264,7 @@
     BOOL ret;
 
     TRACE( "device %s, index %#x, devmode %p, flags %#x\n",
-           debugstr_us(device), index, devmode, flags );
+           debugstr_us(device), (int)index, devmode, (int)flags );
 
     if (!(source = find_source( device ))) return FALSE;
 
@@ -4486,9 +4282,9 @@
 
     if (!ret) WARN( "Failed to query %s display settings.\n", debugstr_us(device) );
     else TRACE( "position %dx%d, resolution %ux%u, frequency %u, depth %u, orientation %#x.\n",
-                devmode->dmPosition.x, devmode->dmPosition.y, devmode->dmPelsWidth,
-                devmode->dmPelsHeight, devmode->dmDisplayFrequency,
-                devmode->dmBitsPerPel, devmode->dmDisplayOrientation );
+                (int)devmode->dmPosition.x, (int)devmode->dmPosition.y, (int)devmode->dmPelsWidth,
+                (int)devmode->dmPelsHeight, (int)devmode->dmDisplayFrequency,
+                (int)devmode->dmBitsPerPel, (int)devmode->dmDisplayOrientation );
     return ret;
 }
 
@@ -4642,7 +4438,7 @@
         monitor_get_info( monitor, info, dpi );
         unlock_display_devices();
 
-        TRACE( "flags %04x, monitor %s, work %s\n", info->dwFlags,
+        TRACE( "flags %04x, monitor %s, work %s\n", (int)info->dwFlags,
                wine_dbgstr_rect(&info->rcMonitor), wine_dbgstr_rect(&info->rcWork));
         return TRUE;
     }
@@ -4682,16 +4478,23 @@
     return info;
 }
 
-UINT monitor_dpi_from_rect( RECT rect, UINT dpi, UINT *raw_dpi )
+UINT monitor_dpi_from_rect( RECT rect, UINT dpi, UINT *raw_dpi_num, UINT *raw_dpi_den )
 {
     struct monitor *monitor;
-    UINT ret = system_dpi, x, y;
+    UINT ret = system_dpi, x, y, d;
 
     if (!lock_display_devices( FALSE )) return 0;
     if ((monitor = get_monitor_from_rect( rect, MONITOR_DEFAULTTONEAREST, dpi, MDT_DEFAULT )))
     {
-        *raw_dpi = monitor_get_dpi( monitor, MDT_RAW_DPI, &x, &y );
         ret = monitor_get_dpi( monitor, MDT_DEFAULT, &x, &y );
+        if (raw_dpi_num && raw_dpi_den)
+        {
+            monitor_virt_to_raw_ratio( monitor, raw_dpi_num, raw_dpi_den );
+            *raw_dpi_num *= ret;
+            d = gcd( *raw_dpi_num, *raw_dpi_den );
+            *raw_dpi_num /= d;
+            *raw_dpi_den /= d;
+        }
     }
     unlock_display_devices();
 
@@ -5940,8 +5743,20 @@
         grab_fullscreen = IS_OPTION_TRUE( buffer[0] );
     if (!get_config_key( hkey, appkey, "Decorated", buffer, sizeof(buffer) ))
         decorated_mode = IS_OPTION_TRUE( buffer[0] );
+    if (!get_config_key( hkey, appkey, "EmulateModelist", buffer, sizeof(buffer) ))
+        emulate_modelist = !IS_OPTION_TRUE( buffer[0] );
     if (!get_config_key( hkey, appkey, "EmulateModeset", buffer, sizeof(buffer) ))
-        emulate_modeset = IS_OPTION_TRUE( buffer[0] );
+        emulate_modeset = !IS_OPTION_TRUE( buffer[0] );
+
+    {
+        const char *s;
+
+        if ((s = getenv( "PROTON_LIMIT_RESOLUTIONS" )))
+        {
+            limit_resolutions = atoi( s );
+            ERR( "HACK: limit_resolutions %u.\n", limit_resolutions );
+        }
+    }
 
 #undef IS_OPTION_TRUE
 
@@ -7338,7 +7153,7 @@
         return FALSE;
     }
 
-    TRACE( "set to %#x\n", context );
+    TRACE( "set to %#x\n", (UINT)context );
     return TRUE;
 }
 
@@ -7361,29 +7176,6 @@
 }
 
 /***********************************************************************
- *	     NtUserGetProcessDefaultLayout    (win32u.@)
- */
-BOOL WINAPI NtUserGetProcessDefaultLayout( ULONG *layout )
-{
-    if (!layout)
-    {
-        RtlSetLastWin32Error( ERROR_NOACCESS );
-        return FALSE;
-    }
-    *layout = process_layout;
-    return TRUE;
-}
-
-/***********************************************************************
- *	     NtUserSetProcessDefaultLayout    (win32u.@)
- */
-BOOL WINAPI NtUserSetProcessDefaultLayout( ULONG layout )
-{
-    process_layout = layout;
-    return TRUE;
-}
-
-/***********************************************************************
  *	     NtUserMessageBeep    (win32u.@)
  */
 BOOL WINAPI NtUserMessageBeep( UINT type )
@@ -7394,16 +7186,6 @@
     return TRUE;
 }
 
-/***********************************************************************
- *	     NtUserSetAdditionalForegroundBoostProcesses    (win32u.@)
- */
-BOOL WINAPI NtUserSetAdditionalForegroundBoostProcesses( HWND hwnd, DWORD count, HANDLE *handles )
-{
-    FIXME( "%p %u %p stub!\n", hwnd, count, handles );
-    RtlSetLastWin32Error( ERROR_CALL_NOT_IMPLEMENTED );
-    return FALSE;
-}
-
 static DWORD exiting_thread_id;
 
 /**********************************************************************
@@ -7425,7 +7207,6 @@
 
     cleanup_imm_thread();
     NtClose( thread_info->server_queue );
-    if (thread_info->idle_event) NtClose( thread_info->idle_event );
     free( thread_info->session_data );
     free( thread_info->mouse_tracking_info );
 
@@ -7457,6 +7238,9 @@
     case NtUserCallNoParam_GetLastInputTime:
         return get_last_input_time();
 
+    case NtUserCallNoParam_GetProcessDefaultLayout:
+        return process_layout;
+
     case NtUserCallNoParam_GetProgmanWindow:
         return HandleToUlong( get_progman_window() );
 
@@ -7480,7 +7264,7 @@
         return 0;
 
     default:
-        FIXME( "invalid code %u\n", code );
+        FIXME( "invalid code %u\n", (int)code );
         return 0;
     }
 }
@@ -7492,6 +7276,9 @@
 {
     switch(code)
     {
+    case NtUserCallOneParam_BeginDeferWindowPos:
+        return HandleToUlong( begin_defer_window_pos( arg ));
+
     case NtUserCallOneParam_CreateCursorIcon:
         return HandleToUlong( alloc_cursoricon_handle( arg ));
 
@@ -7502,6 +7289,12 @@
         thunk_lock_callback = arg;
         return 0;
 
+    case NtUserCallOneParam_GetClipCursor:
+        return get_clip_cursor( (RECT *)arg, get_thread_dpi(), MDT_DEFAULT );
+
+    case NtUserCallOneParam_GetCursorPos:
+        return get_cursor_pos( (POINT *)arg );
+
     case NtUserCallOneParam_GetIconParam:
         return get_icon_param( UlongToHandle(arg) );
 
@@ -7511,6 +7304,9 @@
     case NtUserCallOneParam_GetSysColor:
         return get_sys_color( arg );
 
+    case NtUserCallOneParam_RealizePalette:
+        return realize_palette( UlongToHandle(arg) );
+
     case NtUserCallOneParam_GetPrimaryMonitorRect:
         *(RECT *)arg = get_primary_monitor_rect( 0 );
         return 1;
@@ -7524,6 +7320,16 @@
     case NtUserCallOneParam_GetSystemMetrics:
         return get_system_metrics( arg );
 
+    case NtUserCallOneParam_ReplyMessage:
+        return reply_message_result( arg );
+
+    case NtUserCallOneParam_SetCaretBlinkTime:
+        return set_caret_blink_time( arg );
+
+    case NtUserCallOneParam_SetProcessDefaultLayout:
+        process_layout = arg;
+        return TRUE;
+
     case NtUserCallOneParam_SetKeyboardAutoRepeat:
         return set_keyboard_auto_repeat( arg );
 
@@ -7536,12 +7342,15 @@
     case NtUserCallOneParam_GetAsyncKeyboardState:
         return get_async_keyboard_state( (void *)arg );
 
+    case NtUserCallOneParam_UnregisterTouchWindow:
+        return unregister_touch_window( (HWND)arg );
+
     /* temporary exports */
     case NtUserGetDeskPattern:
         return get_entry( &entry_DESKPATTERN, 256, (WCHAR *)arg );
 
     default:
-        FIXME( "invalid code %u\n", code );
+        FIXME( "invalid code %u\n", (int)code );
         return 0;
     }
 }
@@ -7568,12 +7377,21 @@
     case NtUserCallTwoParam_MonitorFromRect:
         return HandleToUlong( monitor_from_rect( (const RECT *)arg1, arg2, get_thread_dpi() ));
 
+    case NtUserCallTwoParam_RegisterTouchWindow:
+        return register_touch_window( (HWND)arg1, arg2 );
+
+    case NtUserCallTwoParam_SetCaretPos:
+        return set_caret_pos( arg1, arg2 );
+
     case NtUserCallTwoParam_SetIconParam:
         return set_icon_param( UlongToHandle(arg1), UlongToHandle(arg2) );
 
     case NtUserCallTwoParam_SetIMECompositionRect:
         return set_ime_composition_rect( UlongToHandle(arg1), *(const RECT *)arg2 );
 
+    case NtUserCallTwoParam_UnhookWindowsHook:
+        return unhook_windows_hook( arg1, (HOOKPROC)arg2 );
+
     case NtUserCallTwoParam_AdjustWindowRect:
     {
         struct adjust_window_rect_params *params = (void *)arg2;
@@ -7589,7 +7407,7 @@
         return (UINT_PTR)alloc_winproc( (WNDPROC)arg1, arg2 );
 
     default:
-        FIXME( "invalid code %u\n", code );
+        FIXME( "invalid code %u\n", (int)code );
         return 0;
     }
 }
@@ -7792,45 +7610,29 @@
     }
     case DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO:
     {
-        DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO *color_info = (DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO *)packet;
-        struct monitor *monitor;
+        DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO *info = (DISPLAYCONFIG_GET_ADVANCED_COLOR_INFO *)packet;
+        const char *env;
 
         FIXME( "DISPLAYCONFIG_DEVICE_INFO_GET_ADVANCED_COLOR_INFO semi-stub.\n" );
 
-        if (packet->size < sizeof(*color_info))
+        if (packet->size < sizeof(*info))
             return STATUS_INVALID_PARAMETER;
 
-        if (!lock_display_devices( FALSE )) return STATUS_UNSUCCESSFUL;
-
-        LIST_FOR_EACH_ENTRY(monitor, &monitors, struct monitor, entry)
-        {
-            if (color_info->header.id != monitor->output_id) continue;
-            if (memcmp( &color_info->header.adapterId, &monitor->source->gpu->luid,
-                        sizeof(monitor->source->gpu->luid) ))
-                continue;
-
-            if (monitor->hdr_enabled)
-            {
-                color_info->advancedColorSupported = 1;
-                color_info->advancedColorEnabled = 1;
-                color_info->bitsPerColorChannel = 10;
-            }
-            else
-            {
-                color_info->advancedColorSupported = 0;
-                color_info->advancedColorEnabled = 0;
-                color_info->bitsPerColorChannel = 8;
-            }
-            color_info->wideColorEnforced = 0;
-            color_info->advancedColorForceDisabled = 0;
-            color_info->colorEncoding = DISPLAYCONFIG_COLOR_ENCODING_RGB;
-
-            ret = STATUS_SUCCESS;
-            break;
+        info->advancedColorSupported = 0;
+        info->advancedColorEnabled = 0;
+        info->wideColorEnforced = 0;
+        info->advancedColorForceDisabled = 0;
+        info->colorEncoding = DISPLAYCONFIG_COLOR_ENCODING_RGB;
+        info->bitsPerColorChannel = 8;
+        if ((env = getenv("DXVK_HDR")) && *env == '1')
+        {
+            TRACE( "HDR is enabled.\n" );
+            info->advancedColorSupported = 1;
+            info->advancedColorEnabled = 1;
+            info->bitsPerColorChannel = 10;
         }
 
-        unlock_display_devices();
-        return ret;
+        return STATUS_SUCCESS;
     }
     case DISPLAYCONFIG_DEVICE_INFO_SET_TARGET_PERSISTENCE:
     case DISPLAYCONFIG_DEVICE_INFO_GET_TARGET_BASE_TYPE:
@@ -7898,7 +7700,7 @@
         /* give the GDI driver a chance to be notified about new adapter handle */
         if ((status = NtGdiDdDDIOpenAdapterFromLuid( &open_adapter_from_luid )))
         {
-            ERR( "Failed to open adapter %u from LUID, status %#x.\n", idx, status );
+            ERR( "Failed to open adapter %u from LUID, status %#x.\n", idx, (UINT)status );
             break;
         }
 
@@ -8027,26 +7829,5 @@
     }
 
     unlock_display_devices();
-    return found;
-}
-
-/* Find the Vulkan LUID corresponding to a device UUID */
-BOOL get_luid_from_vulkan_uuid( const GUID *uuid, LUID *luid, UINT32 *node_mask )
-{
-    BOOL found = FALSE;
-    struct gpu *gpu;
-
-    if (!lock_display_devices( FALSE )) return FALSE;
-
-    LIST_FOR_EACH_ENTRY( gpu, &gpus, struct gpu, entry )
-    {
-        if (!IsEqualGUID( uuid, &gpu->vulkan_uuid )) continue;
-        *luid = gpu->luid;
-        *node_mask = 1;
-        found = TRUE;
-        break;
-    }
-
-    unlock_display_devices();
     return found;
 }
