diff --git a/dlls/ddraw/tests/ddraw2.c b/dlls/ddraw/tests/ddraw2.c
--- a/dlls/ddraw/tests/ddraw2.c
+++ b/dlls/ddraw/tests/ddraw2.c
@@ -21,8 +21,6 @@
 #include <math.h>
 
 #define COBJMACROS
-#include <stdbool.h>
-#include <stdint.h>
 #include "wine/test.h"
 #include <limits.h>
 #include <math.h>
@@ -30,8 +28,7 @@
 #include "d3dhal.h"
 
 static BOOL is_ddraw64 = sizeof(DWORD) != sizeof(DWORD *);
-static DEVMODEA registry_mode;
-static HWND foreground;
+static DEVMODEW registry_mode;
 
 static HRESULT (WINAPI *pDwmIsCompositionEnabled)(BOOL *);
 
@@ -113,64 +110,6 @@
             DispatchMessageA(&msg);
         diff = time - GetTickCount();
     }
-}
-
-static HWND create_foreground_window(void)
-{
-    for (UINT retries = 5; retries; retries--)
-    {
-        HWND hwnd;
-        BOOL ret;
-
-        hwnd = CreateWindowW(L"static", NULL, WS_POPUP | WS_VISIBLE, 0, 0, 5, 5, NULL, NULL, NULL, NULL);
-        ok(hwnd != NULL, "CreateWindowW failed, error %lu\n", GetLastError());
-        flush_events();
-
-        if (GetForegroundWindow() == hwnd)
-            return hwnd;
-        ret = DestroyWindow(hwnd);
-        ok(ret, "DestroyWindow failed, error %lu\n", GetLastError());
-        flush_events();
-    }
-
-    ok(0, "Failed to create foreground window\n");
-    return NULL;
-}
-
-static LRESULT CALLBACK foreground_window_wndproc(HWND hwnd, UINT msg, WPARAM wparam, LPARAM lparam)
-{
-    if (msg == WM_USER)
-        SetForegroundWindow((HWND)wparam);
-    if (msg == WM_CLOSE)
-        PostQuitMessage(0);
-    return DefWindowProcW(hwnd, msg, wparam, lparam);
-}
-
-static DWORD WINAPI foreground_window_thread(void *arg)
-{
-    HANDLE event = arg;
-    MSG msg;
-
-    foreground = create_foreground_window();
-    SetWindowLongPtrW(foreground, GWLP_WNDPROC, (LONG_PTR)foreground_window_wndproc);
-    SetEvent(event);
-
-    while (GetMessageW(&msg, NULL, 0, 0))
-    {
-        if (msg.message == WM_QUIT) break;
-        TranslateMessage(&msg);
-        DispatchMessageW(&msg);
-    }
-
-    return 0;
-}
-
-static void start_foreground_window_thread(void)
-{
-    HANDLE event = CreateEventA(NULL, FALSE, FALSE, NULL);
-    CloseHandle(CreateThread(NULL, 0, foreground_window_thread, event, 0, NULL));
-    WaitForSingleObject(event, 10000);
-    CloseHandle(event);
 }
 
 static BOOL ddraw_get_identifier(IDirectDraw2 *ddraw, DDDEVICEIDENTIFIER *identifier)
@@ -330,20 +269,20 @@
 }
 
 /* Free original_modes after finished using it */
-static BOOL save_display_modes(DEVMODEA **original_modes, unsigned int *display_count)
+static BOOL save_display_modes(DEVMODEW **original_modes, unsigned int *display_count)
 {
     unsigned int number, size = 2, count = 0, index = 0;
-    DISPLAY_DEVICEA display_device;
-    DEVMODEA *modes, *tmp;
+    DISPLAY_DEVICEW display_device;
+    DEVMODEW *modes, *tmp;
 
     if (!(modes = malloc(size * sizeof(*modes))))
         return FALSE;
 
     display_device.cb = sizeof(display_device);
-    while (EnumDisplayDevicesA(NULL, index++, &display_device, 0))
+    while (EnumDisplayDevicesW(NULL, index++, &display_device, 0))
     {
         /* Skip software devices */
-        if (sscanf(display_device.DeviceName, "\\\\.\\DISPLAY%u", &number) != 1)
+        if (swscanf(display_device.DeviceName, L"\\\\.\\DISPLAY%u", &number) != 1)
             continue;
 
         if (!(display_device.StateFlags & DISPLAY_DEVICE_ATTACHED_TO_DESKTOP))
@@ -362,13 +301,13 @@
 
         memset(&modes[count], 0, sizeof(modes[count]));
         modes[count].dmSize = sizeof(modes[count]);
-        if (!EnumDisplaySettingsA(display_device.DeviceName, ENUM_CURRENT_SETTINGS, &modes[count]))
+        if (!EnumDisplaySettingsW(display_device.DeviceName, ENUM_CURRENT_SETTINGS, &modes[count]))
         {
             free(modes);
             return FALSE;
         }
 
-        strcpy((char *)modes[count++].dmDeviceName, display_device.DeviceName);
+        lstrcpyW(modes[count++].dmDeviceName, display_device.DeviceName);
     }
 
     *original_modes = modes;
@@ -376,19 +315,19 @@
     return TRUE;
 }
 
-static BOOL restore_display_modes(DEVMODEA *modes, unsigned int count)
+static BOOL restore_display_modes(DEVMODEW *modes, unsigned int count)
 {
     unsigned int index;
     LONG ret;
 
     for (index = 0; index < count; ++index)
     {
-        ret = ChangeDisplaySettingsExA((char *)modes[index].dmDeviceName, &modes[index], NULL,
+        ret = ChangeDisplaySettingsExW(modes[index].dmDeviceName, &modes[index], NULL,
                 CDS_UPDATEREGISTRY | CDS_NORESET, NULL);
         if (ret != DISP_CHANGE_SUCCESSFUL)
             return FALSE;
     }
-    ret = ChangeDisplaySettingsExA(NULL, NULL, NULL, 0, NULL);
+    ret = ChangeDisplaySettingsExW(NULL, NULL, NULL, 0, NULL);
     return ret == DISP_CHANGE_SUCCESSFUL;
 }
 
@@ -960,8 +899,6 @@
     window = CreateWindowA("static", "ddraw_test", WS_OVERLAPPEDWINDOW,
             10, 10, 640, 480, 0, 0, 0, 0);
     ShowWindow(window, SW_SHOW);
-    flush_events();
-
     ddraw = create_ddraw();
     ok(!!ddraw, "Failed to create a ddraw object.\n");
 
@@ -1651,7 +1588,7 @@
     if (ddraw) IDirectDraw2_Release(ddraw);
 }
 
-static BOOL compare_mode_rect(const DEVMODEA *mode1, const DEVMODEA *mode2)
+static BOOL compare_mode_rect(const DEVMODEW *mode1, const DEVMODEW *mode2)
 {
     return mode1->dmPosition.x == mode2->dmPosition.x
             && mode1->dmPosition.y == mode2->dmPosition.y
@@ -2774,7 +2711,7 @@
     GetClientRect(window, &r);
     todo_wine ok(!EqualRect(&r, &fullscreen_rect), "Client rect and window rect are equal.\n");
 
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
 
     tmp = GetWindowLongA(window, GWL_STYLE);
@@ -2862,7 +2799,7 @@
     ok(tmp == exstyle, "Expected window extended style %#lx, got %#lx.\n", exstyle, tmp);
 
     ShowWindow(window, SW_SHOW);
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     SetActiveWindow(window);
     ok(GetActiveWindow() == window, "Unexpected active window.\n");
@@ -2934,7 +2871,7 @@
     expected_style = exstyle | WS_EX_TOPMOST;
     todo_wine ok(tmp == expected_style, "Expected window extended style %#lx, got %#lx.\n", expected_style, tmp);
 
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     tmp = GetWindowLongA(window, GWL_STYLE);
     expected_style = style | WS_VISIBLE | WS_MINIMIZE;
@@ -3093,7 +3030,7 @@
 
 static void test_coop_level_mode_set(void)
 {
-    DEVMODEA *original_modes = NULL, devmode, devmode2;
+    DEVMODEW *original_modes = NULL, devmode, devmode2;
     unsigned int display_count = 0;
     IDirectDrawSurface *primary;
     RECT registry_rect, ddraw_rect, user32_rect, r;
@@ -3188,11 +3125,11 @@
 
     memset(&devmode, 0, sizeof(devmode));
     devmode.dmSize = sizeof(devmode);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode, &registry_mode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode, &registry_mode), "Got a different mode.\n");
 
     ret = save_display_modes(&original_modes, &display_count);
@@ -3223,7 +3160,7 @@
     devmode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
     devmode.dmPelsWidth = param.user32_width;
     devmode.dmPelsHeight = param.user32_height;
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
 
     ddraw = create_ddraw();
@@ -3328,7 +3265,7 @@
 
     hr = IDirectDrawSurface_IsLost(primary);
     ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
     hr = IDirectDrawSurface_IsLost(primary);
     ok(hr == DDERR_SURFACELOST, "Got unexpected hr %#lx.\n", hr);
@@ -3344,12 +3281,12 @@
             wine_dbgstr_rect(&r));
 
     expect_messages = exclusive_focus_loss_messages;
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     ok(!expect_messages->message, "Expected message %#x, but didn't receive it.\n", expect_messages->message);
     memset(&devmode, 0, sizeof(devmode));
     devmode.dmSize = sizeof(devmode);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
     ok(ret, "Failed to get display mode.\n");
     ok(devmode.dmPelsWidth == registry_mode.dmPelsWidth
             && devmode.dmPelsHeight == registry_mode.dmPelsHeight, "Got unexpected screen size %lux%lu.\n",
@@ -3362,7 +3299,7 @@
     GetWindowRect(window, &r);
     ok(EqualRect(&r, &ddraw_rect), "Expected %s, got %s.\n", wine_dbgstr_rect(&ddraw_rect),
             wine_dbgstr_rect(&r));
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
     ok(ret, "Failed to get display mode.\n");
     ok(devmode.dmPelsWidth == param.ddraw_width
             && devmode.dmPelsHeight == param.ddraw_height, "Got unexpected screen size %lux%lu.\n",
@@ -3435,7 +3372,7 @@
     expect_messages = NULL;
 
     /* For Wine. */
-    change_ret = ChangeDisplaySettingsA(NULL, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(NULL, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
     flush_events();
 
@@ -3516,7 +3453,7 @@
     devmode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
     devmode.dmPelsWidth = param.user32_width;
     devmode.dmPelsHeight = param.user32_height;
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
     hr = IDirectDrawSurface_IsLost(primary);
     todo_wine ok(hr == DDERR_SURFACELOST, "Got unexpected hr %#lx.\n", hr);
@@ -3629,14 +3566,14 @@
     ok(expect_messages->message == WM_PAINT, "Unexpected WM_PAINT.\n");
     expect_messages = NULL;
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
     ok(ret, "Failed to get display mode.\n");
     ok(devmode.dmPelsWidth == registry_mode.dmPelsWidth
             && devmode.dmPelsHeight == registry_mode.dmPelsHeight,
             "Expected resolution %lux%lu, got %lux%lu.\n",
             registry_mode.dmPelsWidth, registry_mode.dmPelsHeight,
             devmode.dmPelsWidth, devmode.dmPelsHeight);
-    change_ret = ChangeDisplaySettingsA(NULL, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(NULL, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
 
     memset(&ddsd, 0, sizeof(ddsd));
@@ -3713,7 +3650,7 @@
     devmode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
     devmode.dmPelsWidth = param.user32_width;
     devmode.dmPelsHeight = param.user32_height;
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
     hr = IDirectDrawSurface_IsLost(primary);
     todo_wine ok(hr == DDERR_SURFACELOST, "Got unexpected hr %#lx.\n", hr);
@@ -3819,14 +3756,14 @@
     ok(expect_messages->message == WM_PAINT, "Unexpected WM_PAINT.\n");
     expect_messages = NULL;
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
     ok(ret, "Failed to get display mode.\n");
     ok(devmode.dmPelsWidth == registry_mode.dmPelsWidth
             && devmode.dmPelsHeight == registry_mode.dmPelsHeight,
             "Expected resolution %lux%lu, got %lux%lu.\n",
             registry_mode.dmPelsWidth, registry_mode.dmPelsHeight,
             devmode.dmPelsWidth, devmode.dmPelsHeight);
-    change_ret = ChangeDisplaySettingsA(NULL, CDS_FULLSCREEN);
+    change_ret = ChangeDisplaySettingsW(NULL, CDS_FULLSCREEN);
     ok(change_ret == DISP_CHANGE_SUCCESSFUL, "Failed to change display mode, ret %#lx.\n", change_ret);
 
     memset(&ddsd, 0, sizeof(ddsd));
@@ -4008,8 +3945,8 @@
     devmode.dmFields = DM_PELSWIDTH | DM_PELSHEIGHT;
     devmode.dmPelsWidth = param.user32_width;
     devmode.dmPelsHeight = param.user32_height;
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsW failed with %ld.\n", change_ret);
 
     ddraw = create_ddraw();
     ok(!!ddraw, "Failed to create a ddraw object.\n");
@@ -4018,15 +3955,15 @@
 
     memset(&devmode2, 0, sizeof(devmode2));
     devmode2.dmSize = sizeof(devmode2);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &registry_mode), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
 
     /* Test that no mode restorations if no mode changes happened with fullscreen ddraw objects */
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsW failed with %ld.\n", change_ret);
 
     ddraw = create_ddraw();
     ok(!!ddraw, "Failed to create a ddraw object.\n");
@@ -4037,8 +3974,8 @@
     ref = IDirectDraw2_Release(ddraw);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &registry_mode), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
@@ -4050,17 +3987,17 @@
     hr = set_display_mode(ddraw, registry_mode.dmPelsWidth, registry_mode.dmPelsHeight);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsW failed with %ld.\n", change_ret);
 
     ref = IDirectDraw2_Release(ddraw);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &devmode), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
@@ -4071,17 +4008,17 @@
     hr = set_display_mode(ddraw, param.ddraw_width, param.ddraw_height);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsA(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsW(&devmode, CDS_UPDATEREGISTRY | CDS_NORESET);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsW failed with %ld.\n", change_ret);
 
     hr = IDirectDraw2_RestoreDisplayMode(ddraw);
     ok(hr == DD_OK, "RestoreDisplayMode failed, hr %#lx.\n", hr);
 
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &devmode), "Got a different mode.\n");
 
     ref = IDirectDraw2_Release(ddraw);
@@ -4100,9 +4037,9 @@
 
 static void test_coop_level_mode_set_multi(void)
 {
-    DEVMODEA old_devmode, devmode, devmode2, devmode3, *original_modes = NULL;
+    DEVMODEW old_devmode, devmode, devmode2, devmode3, *original_modes = NULL;
     unsigned int mode_idx = 0, display_idx, display_count = 0;
-    char second_monitor_name[CCHDEVICENAME];
+    WCHAR second_monitor_name[CCHDEVICENAME];
     IDirectDraw2 *ddraw1, *ddraw2;
     LONG change_ret;
     UINT w, h;
@@ -4113,11 +4050,11 @@
 
     memset(&devmode, 0, sizeof(devmode));
     devmode.dmSize = sizeof(devmode);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode, &registry_mode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode, &registry_mode), "Got a different mode.\n");
 
     ret = save_display_modes(&original_modes, &display_count);
@@ -4317,18 +4254,18 @@
     {
         if (original_modes[display_idx].dmPosition.x || original_modes[display_idx].dmPosition.y)
         {
-            strcpy(second_monitor_name, (char *)original_modes[display_idx].dmDeviceName);
+            lstrcpyW(second_monitor_name, original_modes[display_idx].dmDeviceName);
             break;
         }
     }
-    ok(strlen(second_monitor_name), "Got an empty second monitor name.\n");
+    ok(lstrlenW(second_monitor_name), "Got an empty second monitor name.\n");
     memset(&old_devmode, 0, sizeof(old_devmode));
     old_devmode.dmSize = sizeof(old_devmode);
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &old_devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &old_devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
 
     devmode = old_devmode;
-    while (EnumDisplaySettingsA(second_monitor_name, mode_idx++, &devmode))
+    while (EnumDisplaySettingsW(second_monitor_name, mode_idx++, &devmode))
     {
         if (devmode.dmPelsWidth != old_devmode.dmPelsWidth
                 || devmode.dmPelsHeight != old_devmode.dmPelsHeight)
@@ -4344,13 +4281,13 @@
     hr = IDirectDraw2_SetCooperativeLevel(ddraw1, window, DDSCL_EXCLUSIVE | DDSCL_FULLSCREEN);
     ok(hr == DD_OK, "SetCooperativeLevel failed, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     memset(&devmode2, 0, sizeof(devmode2));
     devmode2.dmSize = sizeof(devmode2);
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     if (compare_mode_rect(&devmode2, &old_devmode))
     {
         skip("Failed to change display settings of the second monitor.\n");
@@ -4366,8 +4303,8 @@
 
     memset(&devmode3, 0, sizeof(devmode3));
     devmode3.dmSize = sizeof(devmode3);
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode3);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode3);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode3, &devmode2), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
@@ -4379,17 +4316,17 @@
     hr = set_display_mode(ddraw1, 800, 600);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     ref = IDirectDraw2_Release(ddraw1);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
     ret = restore_display_modes(original_modes, display_count);
     ok(ret, "Failed to restore display modes.\n");
@@ -4400,17 +4337,17 @@
     hr = set_display_mode(ddraw1, 800, 600);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     hr = IDirectDraw2_RestoreDisplayMode(ddraw1);
     ok(hr == DD_OK, "RestoreDisplayMode failed, hr %#lx.\n", hr);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
 
     ref = IDirectDraw2_Release(ddraw1);
@@ -4424,20 +4361,20 @@
     hr = set_display_mode(ddraw1, 800, 600);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL,
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL,
             CDS_UPDATEREGISTRY | CDS_NORESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     ref = IDirectDraw2_Release(ddraw1);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(devmode2.dmPelsWidth == devmode.dmPelsWidth && devmode2.dmPelsHeight == devmode.dmPelsHeight,
             "Expected resolution %lux%lu, got %lux%lu.\n", devmode.dmPelsWidth, devmode.dmPelsHeight,
             devmode2.dmPelsWidth, devmode2.dmPelsHeight);
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(devmode2.dmPelsWidth == devmode.dmPelsWidth && devmode2.dmPelsHeight == devmode.dmPelsHeight,
             "Expected resolution %lux%lu, got %lux%lu.\n", devmode.dmPelsWidth, devmode.dmPelsHeight,
             devmode2.dmPelsWidth, devmode2.dmPelsHeight);
@@ -4455,17 +4392,17 @@
     hr = set_display_mode(ddraw2, 640, 480);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     hr = IDirectDraw2_RestoreDisplayMode(ddraw2);
     ok(hr == DD_OK, "RestoreDisplayMode failed, hr %#lx.\n", hr);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
 
     ref = IDirectDraw2_Release(ddraw2);
@@ -4486,17 +4423,17 @@
     hr = set_display_mode(ddraw2, 640, 480);
     ok(hr == DD_OK, "Failed to set display mode, hr %#lx.\n", hr);
 
-    change_ret = ChangeDisplaySettingsExA(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
-    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExA failed with %ld.\n", change_ret);
+    change_ret = ChangeDisplaySettingsExW(second_monitor_name, &devmode, NULL, CDS_RESET, NULL);
+    ok(change_ret == DISP_CHANGE_SUCCESSFUL, "ChangeDisplaySettingsExW failed with %ld.\n", change_ret);
 
     ref = IDirectDraw2_Release(ddraw2);
     ok(!ref, "Unexpected refcount %lu.\n", ref);
 
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_CURRENT_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
-    ret = EnumDisplaySettingsA(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(second_monitor_name, ENUM_REGISTRY_SETTINGS, &devmode2);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
     ok(compare_mode_rect(&devmode2, &old_devmode), "Got a different mode.\n");
 
     ref = IDirectDraw2_Release(ddraw1);
@@ -4987,7 +4924,7 @@
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
 
     /* Exclusive with window not active. */
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     activateapp_testdata.received = FALSE;
     hr = IDirectDraw2_SetCooperativeLevel(ddraw, window, DDSCL_EXCLUSIVE | DDSCL_FULLSCREEN);
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
@@ -4996,7 +4933,7 @@
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
 
     /* Normal with window not active, then exclusive with the same window. */
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     activateapp_testdata.received = FALSE;
     hr = IDirectDraw2_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
@@ -5008,7 +4945,7 @@
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
 
     /* Recursive set of DDSCL_EXCLUSIVE | DDSCL_FULLSCREEN. */
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     activateapp_testdata.received = FALSE;
     activateapp_testdata.ddraw = ddraw;
     activateapp_testdata.window = window;
@@ -5030,7 +4967,7 @@
     ok(SUCCEEDED(hr), "Failed to set cooperative level, hr %#lx.\n", hr);
 
     /* Setting DDSCL_NORMAL with recursive invocation. */
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     activateapp_testdata.received = FALSE;
     activateapp_testdata.ddraw = ddraw;
     activateapp_testdata.window = window;
@@ -9319,7 +9256,7 @@
         ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
     }
 
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     hr = IDirectDrawSurface_IsLost(surface);
     ok(hr == DDERR_SURFACELOST, "Got unexpected hr %#lx.\n", hr);
@@ -9402,7 +9339,7 @@
     hr = IDirectDrawSurface_IsLost(surface);
     ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
 
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     hr = IDirectDrawSurface_IsLost(surface);
     ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
@@ -14946,7 +14883,7 @@
     hr = IDirectDraw2_CreateSurface(killfocus_ddraw, &surface_desc, &killfocus_surface, NULL);
     ok(SUCCEEDED(hr), "Failed to create surface, hr %#lx.\n", hr);
 
-    SetForegroundWindow(foreground);
+    SetForegroundWindow(GetDesktopWindow());
     ok(!killfocus_ddraw, "WM_KILLFOCUS was not received.\n");
 
     DestroyWindow(window);
@@ -15335,59 +15272,47 @@
 
 static void test_caps(void)
 {
+    DWORD caps_never, caps_always, caps_hal;
     DDCAPS hal_caps, hel_caps;
     IDirectDraw2 *ddraw;
     IDirectDraw *ddraw1;
-    HWND window;
     HRESULT hr;
     BOOL no3d;
 
-    static const struct
-    {
-        unsigned int depth;
-        DWORD flag;
-    }
-    depth_caps[] =
-    {
-        {16, DDBD_16},
-        {24, DDBD_24},
-        {32, DDBD_32},
-    };
-
-    static const DWORD caps_hel = DDSCAPS_FLIP
-            | DDSCAPS_OFFSCREENPLAIN
-            | DDSCAPS_PALETTE
-            | DDSCAPS_PRIMARYSURFACE
-            | DDSCAPS_TEXTURE
-            | DDSCAPS_ZBUFFER
-            | DDSCAPS_MIPMAP;
-
-    static const DWORD caps_never = DDSCAPS_RESERVED1
+    caps_never = DDSCAPS_RESERVED1
+            | DDSCAPS_ALPHA
             | DDSCAPS_PRIMARYSURFACELEFT
+            | DDSCAPS_SYSTEMMEMORY
+            | DDSCAPS_VISIBLE
             | DDSCAPS_WRITEONLY
             | DDSCAPS_LIVEVIDEO
+            | DDSCAPS_HWCODEC
+            | DDSCAPS_MODEX
             | DDSCAPS_RESERVED2
             | 0x01000000u
             | 0x02000000u
             | DDSCAPS_ALLOCONLOAD
             | DDSCAPS_VIDEOPORT
+            | DDSCAPS_STANDARDVGAMODE
             | DDSCAPS_OPTIMIZED;
 
-    static const DWORD caps_always = DDSCAPS_BACKBUFFER
-            | DDSCAPS_FLIP
+    caps_always = DDSCAPS_FLIP
             | DDSCAPS_OFFSCREENPLAIN
             | DDSCAPS_PRIMARYSURFACE
             | DDSCAPS_TEXTURE
+            | DDSCAPS_ZBUFFER
+            | DDSCAPS_MIPMAP;
+
+    caps_hal = DDSCAPS_BACKBUFFER
+            | DDSCAPS_COMPLEX
+            | DDSCAPS_FRONTBUFFER
             | DDSCAPS_3DDEVICE
-            | DDSCAPS_ZBUFFER
             | DDSCAPS_VIDEOMEMORY
-            | DDSCAPS_MIPMAP;
-
-    window = create_window();
+            | DDSCAPS_LOCALVIDMEM
+            | DDSCAPS_NONLOCALVIDMEM;
+
     ddraw = create_ddraw();
     ok(!!ddraw, "Failed to create a ddraw object.\n");
-    hr = IDirectDraw_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
-    ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
 
     memset(&hal_caps, 0, sizeof(hal_caps));
     memset(&hel_caps, 0, sizeof(hel_caps));
@@ -15406,48 +15331,14 @@
     if (hal_caps.ddsCaps.dwCaps)
     {
         ok(!(hal_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
-        todo_wine_if(no3d) ok(!(~hal_caps.ddsCaps.dwCaps & caps_always),
+        ok(!(~hal_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
+        todo_wine_if(no3d) ok(!(~hal_caps.ddsCaps.dwCaps & caps_hal),
                 "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
     }
-    todo_wine ok(hel_caps.ddsCaps.dwCaps == caps_hel, "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
-
-    ok(!(hal_caps.dwZBufferBitDepths & ~(DDBD_16 | DDBD_24 | DDBD_32)),
-            "Got HAL depth caps %#lx.\n", hal_caps.dwZBufferBitDepths);
-    todo_wine ok(hel_caps.dwZBufferBitDepths == DDBD_16, "Got HEL depth caps %#lx.\n", hel_caps.dwZBufferBitDepths);
-
-    for (unsigned int i = 0; i < ARRAY_SIZE(depth_caps); ++i)
-    {
-        bool supported = (hal_caps.dwZBufferBitDepths & depth_caps[i].flag);
-        IDirectDrawSurface *surface;
-        DDSURFACEDESC desc =
-        {
-            .dwSize = sizeof(DDSURFACEDESC),
-            .dwFlags = DDSD_CAPS | DDSD_ZBUFFERBITDEPTH | DDSD_WIDTH | DDSD_HEIGHT,
-            .ddsCaps.dwCaps = DDSCAPS_ZBUFFER,
-            .dwZBufferBitDepth = depth_caps[i].depth,
-            .dwWidth = 64,
-            .dwHeight = 64,
-        };
-
-        winetest_push_context("depth %u", depth_caps[i].depth);
-
-        /* dwZBufferBitDepths sometimes reports false negatives,
-         * but it has not been known to report false positives. */
-        hr = IDirectDraw2_CreateSurface(ddraw, &desc, &surface, NULL);
-        ok(hr == S_OK || (!supported && hr == DDERR_INVALIDPIXELFORMAT), "Got hr %#lx.\n", hr);
-
-        if (hr == S_OK)
-        {
-            hr = IDirectDrawSurface_GetSurfaceDesc(surface, &desc);
-            ok(hr == S_OK, "Got hr %#lx.\n", hr);
-            ok(desc.ddsCaps.dwCaps == (DDSCAPS_VIDEOMEMORY | DDSCAPS_LOCALVIDMEM | DDSCAPS_ZBUFFER)
-                    || (!supported && (desc.ddsCaps.dwCaps == (DDSCAPS_SYSTEMMEMORY | DDSCAPS_ZBUFFER))),
-                    "Got caps %#lx.\n", desc.ddsCaps.dwCaps);
-            IDirectDrawSurface_Release(surface);
-        }
-
-        winetest_pop_context();
-    }
+    ok(!(hel_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+    ok(!(~hel_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+    todo_wine_if(!no3d) ok(!(hel_caps.ddsCaps.dwCaps & caps_hal),
+            "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
 
     IDirectDraw2_Release(ddraw);
 
@@ -15473,11 +15364,20 @@
                 hel_caps.ddsOldCaps.dwCaps, hel_caps.ddsCaps.dwCaps);
 
         ok(!(hal_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
-        todo_wine_if(no3d) ok(!(~hal_caps.ddsCaps.dwCaps & caps_always),
+        ok(!(~hal_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
+        todo_wine_if(no3d) ok(!(~hal_caps.ddsCaps.dwCaps & caps_hal),
                 "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
-        todo_wine ok(!(hel_caps.ddsCaps.dwCaps & ~(DDSCAPS_STANDARDVGAMODE | DDSCAPS_MODEX))
-                || broken(hel_caps.ddsCaps.dwCaps == caps_hel), /* Windows 8+ */
-                "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+        if (is_ddraw64)
+        {
+            ok(!(hel_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+            ok(!(~hel_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+            todo_wine_if(!no3d) ok(!(hel_caps.ddsCaps.dwCaps & caps_hal),
+                    "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+        }
+        else
+        {
+            todo_wine ok(!hel_caps.ddsCaps.dwCaps, "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+        }
 
         IDirectDraw2_Release(ddraw);
     }
@@ -15489,8 +15389,6 @@
         hr = IDirectDraw_QueryInterface(ddraw1, &IID_IDirectDraw2, (void **)&ddraw);
         ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
         IDirectDraw_Release(ddraw1);
-        hr = IDirectDraw2_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
-        ok(hr == DD_OK, "Got unexpected hr %#lx.\n", hr);
 
         memset(&hal_caps, 0, sizeof(hal_caps));
         memset(&hel_caps, 0, sizeof(hel_caps));
@@ -15505,48 +15403,14 @@
                 "Got unexpected caps %#lx, expected %#lx.\n",
                 hel_caps.ddsOldCaps.dwCaps, hel_caps.ddsCaps.dwCaps);
 
-        todo_wine ok(!(hal_caps.ddsCaps.dwCaps & ~(DDSCAPS_STANDARDVGAMODE | DDSCAPS_MODEX)),
-                "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
-        todo_wine ok(hel_caps.ddsCaps.dwCaps == caps_hel, "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
-
-        for (unsigned int i = 0; i < ARRAY_SIZE(depth_caps); ++i)
-        {
-            IDirectDrawSurface *surface;
-            DDSURFACEDESC desc =
-            {
-                .dwSize = sizeof(DDSURFACEDESC),
-                .dwFlags = DDSD_CAPS | DDSD_ZBUFFERBITDEPTH | DDSD_WIDTH | DDSD_HEIGHT,
-                .ddsCaps.dwCaps = DDSCAPS_ZBUFFER,
-                .dwZBufferBitDepth = depth_caps[i].depth,
-                .dwWidth = 64,
-                .dwHeight = 64,
-            };
-
-            winetest_push_context("depth %u", depth_caps[i].depth);
-
-            hr = IDirectDraw2_CreateSurface(ddraw, &desc, &surface, NULL);
-            if (depth_caps[i].depth == 16 || depth_caps[i].depth == 32)
-                ok(hr == S_OK, "Got hr %#lx.\n", hr);
-            else
-                todo_wine ok(hr == DDERR_INVALIDPIXELFORMAT, "Got hr %#lx.\n", hr);
-
-            if (hr == S_OK)
-            {
-                hr = IDirectDrawSurface_GetSurfaceDesc(surface, &desc);
-                ok(hr == S_OK, "Got hr %#lx.\n", hr);
-                todo_wine_if (depth_caps[i].depth != 32)
-                    ok(desc.ddsCaps.dwCaps == (DDSCAPS_SYSTEMMEMORY | DDSCAPS_ZBUFFER),
-                            "Got caps %#lx.\n", desc.ddsCaps.dwCaps);
-                IDirectDrawSurface_Release(surface);
-            }
-
-            winetest_pop_context();
-        }
+        todo_wine ok(!hal_caps.ddsCaps.dwCaps, "Got unexpected caps %#lx.\n", hal_caps.ddsCaps.dwCaps);
+        ok(!(hel_caps.ddsCaps.dwCaps & caps_never), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+        ok(!(~hel_caps.ddsCaps.dwCaps & caps_always), "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
+        todo_wine_if(!no3d) ok(!(hel_caps.ddsCaps.dwCaps & caps_hal),
+                "Got unexpected caps %#lx.\n", hel_caps.ddsCaps.dwCaps);
 
         IDirectDraw2_Release(ddraw);
     }
-
-    DestroyWindow(window);
 }
 
 static void test_d32_support(void)
@@ -15581,6 +15445,8 @@
     ok((surface_desc.dwFlags & DDSD_ZBUFFERBITDEPTH), "Got unexpected flags %#lx.\n", surface_desc.dwFlags);
     ok(surface_desc.dwZBufferBitDepth == 32,
             "Got unexpected dwZBufferBitDepth %lu.\n", surface_desc.dwZBufferBitDepth);
+    ok(!(surface_desc.ddsCaps.dwCaps & DDSCAPS_VIDEOMEMORY),
+            "Got unexpected surface caps %#lx.\n", surface_desc.ddsCaps.dwCaps);
     IDirectDrawSurface_Release(surface);
 
     refcount = IDirectDraw2_Release(ddraw);
@@ -15813,7 +15679,7 @@
     ret = SetWindowPos(window, 0, monitor_rect->left, monitor_rect->top, 0, 0,
             SWP_NOZORDER | SWP_NOSIZE);
     ok(ret, "SetWindowPos failed, error %lu.\n", GetLastError());
-    ret = SetForegroundWindow(foreground);
+    ret = SetForegroundWindow(GetDesktopWindow());
     ok(ret, "Failed to set foreground window.\n");
     flush_events();
     ret = ShowWindow(window, SW_RESTORE);
@@ -15882,7 +15748,7 @@
     struct find_different_mode_param param;
     DDSURFACEDESC surface_desc;
     IDirectDraw2 *ddraw;
-    DEVMODEA devmode;
+    DEVMODEW devmode;
     HWND window;
     HRESULT hr;
     BOOL ret;
@@ -15899,8 +15765,8 @@
 
     memset(&devmode, 0, sizeof(devmode));
     devmode.dmSize = sizeof(devmode);
-    ret = EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &devmode);
-    ok(ret, "EnumDisplaySettingsA failed, error %lu.\n", GetLastError());
+    ret = EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &devmode);
+    ok(ret, "EnumDisplaySettingsW failed, error %lu.\n", GetLastError());
 
     surface_desc.dwSize = sizeof(surface_desc);
     hr = IDirectDraw2_GetDisplayMode(ddraw, &surface_desc);
@@ -16517,216 +16383,162 @@
     DestroyWindow(window);
 }
 
-static unsigned int enum_devices_index;
-
 static HRESULT WINAPI test_enum_devices_caps_callback(GUID *guid, char *device_desc,
         char *device_name, D3DDEVICEDESC *hal, D3DDEVICEDESC *hel, void *ctx)
 {
-    static const D3DLIGHTINGCAPS empty_lighting_caps = {.dwSize = sizeof(D3DLIGHTINGCAPS)};
-    static const D3DPRIMCAPS empty_primitive_caps = {.dwSize = sizeof(D3DPRIMCAPS)};
-    IDirectDraw2 *ddraw = ctx;
-
-    ok(hal->dwSize == offsetof(D3DDEVICEDESC, dwMaxTextureRepeat), "Got size %lu.\n", hal->dwSize);
-    ok(hel->dwSize == offsetof(D3DDEVICEDESC, dwMaxTextureRepeat), "Got size %lu.\n", hel->dwSize);
-    ok(hal->dtcTransformCaps.dwSize == sizeof(D3DTRANSFORMCAPS)
-            || (ddraw_is_warp(ddraw) && !hal->dtcTransformCaps.dwSize),
-            "Got transform caps size %lu.\n", hal->dtcTransformCaps.dwSize);
-    ok(hel->dtcTransformCaps.dwSize == sizeof(D3DTRANSFORMCAPS),
-            "Got transform caps size %lu.\n", hel->dtcTransformCaps.dwSize);
-    ok(hal->dlcLightingCaps.dwSize == sizeof(D3DLIGHTINGCAPS)
-            || (ddraw_is_warp(ddraw) && !hal->dlcLightingCaps.dwSize),
-            "Got lighting caps size %lu.\n", hal->dlcLightingCaps.dwSize);
-    ok(hel->dlcLightingCaps.dwSize == sizeof(D3DLIGHTINGCAPS),
-            "Got lighting caps size %lu.\n", hel->dlcLightingCaps.dwSize);
-    ok(hal->dpcLineCaps.dwSize == sizeof(D3DPRIMCAPS),
-            "Got line caps size %lu.\n", hal->dpcLineCaps.dwSize);
-    ok(hel->dpcLineCaps.dwSize == sizeof(D3DPRIMCAPS),
-            "Got line caps size %lu.\n", hel->dpcLineCaps.dwSize);
-    ok(hal->dpcTriCaps.dwSize == sizeof(D3DPRIMCAPS),
-            "Got triangle caps size %lu.\n", hal->dpcTriCaps.dwSize);
-    ok(hel->dpcTriCaps.dwSize == sizeof(D3DPRIMCAPS),
-            "Got triangle caps size %lu.\n", hel->dpcTriCaps.dwSize);
-
-    if (!IsEqualGUID(guid, &IID_IDirect3DHALDevice))
-    {
-        ok(!hal->dwFlags, "Got HAL flags %#lx.\n", hal->dwFlags);
-        ok(!hal->dcmColorModel, "Got color model %#lx.\n", hal->dcmColorModel);
-        todo_wine ok(!hal->dwDevCaps, "Got device caps %#lx.\n", hal->dwDevCaps);
-        todo_wine ok(!hal->dtcTransformCaps.dwCaps, "Got transform caps %#lx.\n", hal->dtcTransformCaps.dwCaps);
-        todo_wine ok(!hal->bClipping, "Got clipping %#x.\n", hal->bClipping);
-        todo_wine ok(!memcmp(&hal->dlcLightingCaps, &empty_lighting_caps, sizeof(D3DLIGHTINGCAPS)),
-                "Lighting caps didn't match.\n");
-        todo_wine ok(!memcmp(&hal->dpcLineCaps, &empty_primitive_caps, sizeof(D3DPRIMCAPS)), "Line caps didn't match.\n");
-        todo_wine ok(!memcmp(&hal->dpcTriCaps, &empty_primitive_caps, sizeof(D3DPRIMCAPS)), "Triangle caps didn't match.\n");
-        todo_wine ok(!hal->dwDeviceRenderBitDepth, "Got colour depth %#lx.\n", hal->dwDeviceRenderBitDepth);
-        todo_wine ok(!hal->dwDeviceZBufferBitDepth, "Got Z depth %#lx.\n", hal->dwDeviceZBufferBitDepth);
-        ok(!hal->dwMaxBufferSize, "Got max buffer size %lu.\n", hal->dwMaxBufferSize);
-        todo_wine ok(!hal->dwMaxVertexCount, "Got max vertex count %lu.\n", hal->dwMaxVertexCount);
-        todo_wine ok(!hal->dwMinTextureWidth, "Got min texture width %lu.\n", hal->dwMinTextureWidth);
-        todo_wine ok(!hal->dwMinTextureHeight, "Got min texture height %lu.\n", hal->dwMinTextureHeight);
-        todo_wine ok(!hal->dwMaxTextureWidth, "Got max texture width %lu.\n", hal->dwMaxTextureWidth);
-        todo_wine ok(!hal->dwMaxTextureHeight, "Got max texture height %lu.\n", hal->dwMaxTextureHeight);
-        todo_wine ok(!hal->dwMinStippleWidth, "Got min stipple width %lu.\n", hal->dwMinStippleWidth);
-        todo_wine ok(!hal->dwMinStippleHeight, "Got min stipple height %lu.\n", hal->dwMinStippleHeight);
-        todo_wine ok(!hal->dwMaxStippleWidth, "Got max stipple width %lu.\n", hal->dwMaxStippleWidth);
-        todo_wine ok(!hal->dwMaxStippleHeight, "Got max stipple height %lu.\n", hal->dwMaxStippleHeight);
-    }
-
     if(IsEqualGUID(&IID_IDirect3DRGBDevice, guid))
     {
-        static const DWORD hel_flags = D3DDD_COLORMODEL
-                | D3DDD_DEVCAPS
-                | D3DDD_TRANSFORMCAPS
-                | D3DDD_LIGHTINGCAPS
-                | D3DDD_BCLIPPING
-                | D3DDD_TRICAPS
-                | D3DDD_DEVICERENDERBITDEPTH
-                | D3DDD_DEVICEZBUFFERBITDEPTH
-                | D3DDD_MAXBUFFERSIZE
-                | D3DDD_MAXVERTEXCOUNT;
-
-        static const DWORD device_caps = D3DDEVCAPS_FLOATTLVERTEX
-                | D3DDEVCAPS_SORTINCREASINGZ
-                | D3DDEVCAPS_SORTEXACT
-                | D3DDEVCAPS_EXECUTESYSTEMMEMORY
-                | D3DDEVCAPS_TLVERTEXSYSTEMMEMORY
-                | D3DDEVCAPS_TEXTURESYSTEMMEMORY
-                | D3DDEVCAPS_DRAWPRIMTLVERTEX;
-
-        todo_wine ok(enum_devices_index == 1, "Expected index %u.\n", enum_devices_index);
-        ok(!strcmp(device_name, "RGB Emulation"), "Got name %s.\n", debugstr_a(device_name));
-
-        todo_wine ok(hel->dwFlags == hel_flags, "Got HEL flags %#lx.\n", hel->dwFlags);
-        ok(hel->dcmColorModel == D3DCOLOR_RGB, "Got color model %#lx.\n", hel->dcmColorModel);
-        todo_wine ok(hel->dwDevCaps == device_caps, "Got device caps %#lx.\n", hel->dwDevCaps);
-
+        ok((hal->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2) == 0,
+           "RGB Device hal line caps has D3DPTEXTURECAPS_POW2 flag set\n");
+        ok((hal->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2) == 0,
+           "RGB Device hal tri caps has D3DPTEXTURECAPS_POW2 flag set\n");
         ok(hel->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2,
            "RGB Device hel line caps does not have D3DPTEXTURECAPS_POW2 flag set\n");
         ok(hel->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2,
            "RGB Device hel tri caps does not have D3DPTEXTURECAPS_POW2 flag set\n");
 
+        ok((hal->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE) == 0,
+           "RGB Device hal line caps has D3DPTEXTURECAPS_PERSPECTIVE set\n");
+        ok((hal->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE) == 0,
+           "RGB Device hal tri caps has D3DPTEXTURECAPS_PERSPECTIVE set\n");
         ok(hel->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE,
            "RGB Device hel tri caps does not have D3DPTEXTURECAPS_PERSPECTIVE set\n");
         ok(hel->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE,
            "RGB Device hel tri caps does not have D3DPTEXTURECAPS_PERSPECTIVE set\n");
 
+        ok(hal->dcmColorModel == 0, "RGB Device hal caps has colormodel %lu\n", hal->dcmColorModel);
+        ok(hel->dcmColorModel == D3DCOLOR_RGB, "RGB Device hel caps has colormodel %lu\n", hel->dcmColorModel);
+
+        ok(hal->dwFlags == 0, "RGB Device hal caps has hardware flags %#lx\n", hal->dwFlags);
+        ok(hel->dwFlags != 0, "RGB Device hel caps has hardware flags %#lx\n", hel->dwFlags);
+
+        ok((hal->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
+           "RGB Device hal device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
            "RGB Device hel device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
+        ok((hal->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
+           "RGB Device hal device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
            "RGB Device hel device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
+        ok((hal->dwDevCaps & D3DDEVCAPS_HWRASTERIZATION) == 0,
+           "RGB Device hal device caps has D3DDEVCAPS_HWRASTERIZATION set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_HWRASTERIZATION) == 0,
            "RGB Device hel device caps has D3DDEVCAPS_HWRASTERIZATION set\n");
     }
     else if(IsEqualGUID(&IID_IDirect3DHALDevice, guid))
     {
-        static const DWORD hel_flags = D3DDD_COLORMODEL
-                | D3DDD_DEVCAPS
-                | D3DDD_TRANSFORMCAPS
-                | D3DDD_LIGHTINGCAPS
-                | D3DDD_BCLIPPING;
-
-        todo_wine ok(enum_devices_index == 2, "Expected index %u.\n", enum_devices_index);
-        ok(!strcmp(device_name, "Direct3D HAL"), "Got name %s.\n", debugstr_a(device_name));
-
         ok(hal->dcmColorModel == D3DCOLOR_RGB, "HAL Device hal caps has colormodel %lu\n", hel->dcmColorModel);
+        ok(hel->dcmColorModel == 0, "HAL Device hel caps has colormodel %lu\n", hel->dcmColorModel);
 
         ok(hal->dwFlags != 0, "HAL Device hal caps has hardware flags %#lx\n", hal->dwFlags);
-
+        ok(hel->dwFlags != 0, "HAL Device hel caps has hardware flags %#lx\n", hel->dwFlags);
+
+        ok(hal->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT,
+           "HAL Device hal device caps does not have D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
+        ok((hel->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
+           "RGB Device hel device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
         ok(hal->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX,
            "HAL Device hal device caps does not have D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
-
-        todo_wine ok(hel->dwFlags == hel_flags, "Got HEL flags %#lx.\n", hel->dwFlags);
-        ok(!hel->dcmColorModel, "Got color model %#lx.\n", hel->dcmColorModel);
-        todo_wine ok(hel->dwDevCaps == D3DDEVCAPS_FLOATTLVERTEX, "Got device caps %#lx.\n", hel->dwDevCaps);
-        ok(hel->dtcTransformCaps.dwCaps == D3DTRANSFORMCAPS_CLIP,
-                "Got transform caps %#lx.\n", hel->dtcTransformCaps.dwCaps);
-        ok(hel->bClipping == TRUE, "Got clipping %#x.\n", hel->bClipping);
-        ok(hel->dlcLightingCaps.dwCaps == (D3DLIGHTCAPS_POINT | D3DLIGHTCAPS_SPOT
-                | D3DLIGHTCAPS_DIRECTIONAL | D3DLIGHTCAPS_PARALLELPOINT),
-                "Got lighting caps %#lx.\n", hel->dlcLightingCaps.dwCaps);
-        ok(hel->dlcLightingCaps.dwLightingModel == D3DLIGHTINGMODEL_RGB,
-                "Got lighting model %#lx.\n", hel->dlcLightingCaps.dwLightingModel);
-        todo_wine ok(!hel->dlcLightingCaps.dwNumLights,
-                "Got light count %lu.\n", hel->dlcLightingCaps.dwNumLights);
-        todo_wine ok(!memcmp(&hel->dpcLineCaps, &empty_primitive_caps, sizeof(D3DPRIMCAPS)), "Line caps didn't match.\n");
-        todo_wine ok(!memcmp(&hel->dpcTriCaps, &empty_primitive_caps, sizeof(D3DPRIMCAPS)), "Triangle caps didn't match.\n");
-        todo_wine ok(!hel->dwDeviceRenderBitDepth, "Got colour depth %#lx.\n", hel->dwDeviceRenderBitDepth);
-        todo_wine ok(!hel->dwDeviceZBufferBitDepth, "Got Z depth %#lx.\n", hel->dwDeviceZBufferBitDepth);
-        ok(!hel->dwMaxBufferSize, "Got max buffer size %lu.\n", hel->dwMaxBufferSize);
-        ok(hel->dwMaxVertexCount == hal->dwMaxVertexCount, "Got HAL max vertex count %lu, HEL %lu.\n",
-                hal->dwMaxVertexCount, hel->dwMaxVertexCount);
-        todo_wine ok(!hel->dwMinTextureWidth, "Got min texture width %lu.\n", hel->dwMinTextureWidth);
-        todo_wine ok(!hel->dwMinTextureHeight, "Got min texture height %lu.\n", hel->dwMinTextureHeight);
-        todo_wine ok(!hel->dwMaxTextureWidth, "Got max texture width %lu.\n", hel->dwMaxTextureWidth);
-        todo_wine ok(!hel->dwMaxTextureHeight, "Got max texture height %lu.\n", hel->dwMaxTextureHeight);
-        todo_wine ok(!hel->dwMinStippleWidth, "Got min stipple width %lu.\n", hel->dwMinStippleWidth);
-        todo_wine ok(!hel->dwMinStippleHeight, "Got min stipple height %lu.\n", hel->dwMinStippleHeight);
-        todo_wine ok(!hel->dwMaxStippleWidth, "Got max stipple width %lu.\n", hel->dwMaxStippleWidth);
-        todo_wine ok(!hel->dwMaxStippleHeight, "Got max stipple height %lu.\n", hel->dwMaxStippleHeight);
+        ok((hel->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
+           "RGB Device hel device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
     }
     else if(IsEqualGUID(&IID_IDirect3DRefDevice, guid))
     {
+        ok((hal->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2) == 0,
+           "REF Device hal line caps has D3DPTEXTURECAPS_POW2 flag set\n");
+        ok((hal->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2) == 0,
+           "REF Device hal tri caps has D3DPTEXTURECAPS_POW2 flag set\n");
         ok(hel->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2,
            "REF Device hel line caps does not have D3DPTEXTURECAPS_POW2 flag set\n");
         ok(hel->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2,
            "REF Device hel tri caps does not have D3DPTEXTURECAPS_POW2 flag set\n");
 
+        ok((hal->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE) == 0,
+           "REF Device hal line caps has D3DPTEXTURECAPS_PERSPECTIVE set\n");
+        ok((hal->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE) == 0,
+           "REF Device hal tri caps has D3DPTEXTURECAPS_PERSPECTIVE set\n");
         ok(hel->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE,
            "REF Device hel tri caps does not have D3DPTEXTURECAPS_PERSPECTIVE set\n");
         ok(hel->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE,
            "REF Device hel tri caps does not have D3DPTEXTURECAPS_PERSPECTIVE set\n");
 
+        ok((hal->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
+           "REF Device hal device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
            "REF Device hel device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
+        ok((hal->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
+           "REF Device hal device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
            "REF Device hel device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
     }
     else if(IsEqualGUID(&IID_IDirect3DRampDevice, guid))
     {
-        ok(enum_devices_index == 0, "Expected index %u.\n", enum_devices_index);
-        ok(!strcmp(device_name, "Ramp Emulation"), "Got name %s.\n", debugstr_a(device_name));
-
+        ok((hal->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2) == 0,
+           "Ramp Device hal line caps has D3DPTEXTURECAPS_POW2 flag set\n");
+        ok((hal->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2) == 0,
+           "Ramp Device hal tri caps has D3DPTEXTURECAPS_POW2 flag set\n");
         ok(hel->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2,
            "Ramp Device hel line caps does not have D3DPTEXTURECAPS_POW2 flag set\n");
         ok(hel->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2,
            "Ramp Device hel tri caps does not have D3DPTEXTURECAPS_POW2 flag set\n");
 
+        ok((hal->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE) == 0,
+           "Ramp Device hal line caps has D3DPTEXTURECAPS_PERSPECTIVE set\n");
+        ok((hal->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE) == 0,
+           "Ramp Device hal tri caps has D3DPTEXTURECAPS_PERSPECTIVE set\n");
         ok(hel->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE,
            "Ramp Device hel tri caps does not have D3DPTEXTURECAPS_PERSPECTIVE set\n");
         ok(hel->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE,
            "Ramp Device hel tri caps does not have D3DPTEXTURECAPS_PERSPECTIVE set\n");
 
+        ok(hal->dcmColorModel == 0, "Ramp Device hal caps has colormodel %lu\n", hal->dcmColorModel);
         ok(hel->dcmColorModel == D3DCOLOR_MONO, "Ramp Device hel caps has colormodel %lu\n",
            hel->dcmColorModel);
 
+        ok(hal->dwFlags == 0, "Ramp Device hal caps has hardware flags %#lx\n", hal->dwFlags);
         ok(hel->dwFlags != 0, "Ramp Device hel caps has hardware flags %#lx\n", hel->dwFlags);
 
+        ok((hal->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
+           "Ramp Device hal device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
            "Ramp Device hel device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
+        ok((hal->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
+           "Ramp Device hal device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
            "Ramp Device hel device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
     }
     else if(IsEqualGUID(&IID_IDirect3DMMXDevice, guid))
     {
-        ok(enum_devices_index == 3, "Expected index %u.\n", enum_devices_index);
-        ok(!strcmp(device_name, "MMX Emulation"), "Got name %s.\n", debugstr_a(device_name));
-
+        ok((hal->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2) == 0,
+           "MMX Device hal line caps has D3DPTEXTURECAPS_POW2 flag set\n");
+        ok((hal->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2) == 0,
+           "MMX Device hal tri caps has D3DPTEXTURECAPS_POW2 flag set\n");
         ok(hel->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2,
            "MMX Device hel line caps does not have D3DPTEXTURECAPS_POW2 flag set\n");
         ok(hel->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_POW2,
            "MMX Device hel tri caps does not have D3DPTEXTURECAPS_POW2 flag set\n");
 
+        ok((hal->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE) == 0,
+           "MMX Device hal line caps has D3DPTEXTURECAPS_PERSPECTIVE set\n");
+        ok((hal->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE) == 0,
+           "MMX Device hal tri caps has D3DPTEXTURECAPS_PERSPECTIVE set\n");
         ok(hel->dpcLineCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE,
            "MMX Device hel tri caps does not have D3DPTEXTURECAPS_PERSPECTIVE set\n");
         ok(hel->dpcTriCaps.dwTextureCaps & D3DPTEXTURECAPS_PERSPECTIVE,
            "MMX Device hel tri caps does not have D3DPTEXTURECAPS_PERSPECTIVE set\n");
 
+        ok(hal->dcmColorModel == 0, "MMX Device hal caps has colormodel %lu\n", hal->dcmColorModel);
         ok(hel->dcmColorModel == D3DCOLOR_RGB, "MMX Device hel caps has colormodel %lu\n", hel->dcmColorModel);
 
+        ok(hal->dwFlags == 0, "MMX Device hal caps has hardware flags %#lx\n", hal->dwFlags);
         ok(hel->dwFlags != 0, "MMX Device hel caps has hardware flags %#lx\n", hel->dwFlags);
 
+        ok((hal->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
+           "MMX Device hal device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_HWTRANSFORMANDLIGHT) == 0,
            "MMX Device hel device caps has D3DDEVCAPS_HWTRANSFORMANDLIGHT set\n");
+        ok((hal->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
+           "MMX Device hal device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
         ok((hel->dwDevCaps & D3DDEVCAPS_DRAWPRIMITIVES2EX) == 0,
            "MMX Device hel device caps has D3DDEVCAPS_DRAWPRIMITIVES2EX set\n");
     }
@@ -16751,7 +16563,6 @@
             trace("hel tri does NOT have pow2 set\n");
     }
 
-    ++enum_devices_index;
     return DDENUMRET_OK;
 }
 
@@ -16776,7 +16587,7 @@
     hr = IDirect3D2_EnumDevices(d3d, NULL, NULL);
     ok(hr == DDERR_INVALIDPARAMS, "Got hr %#lx.\n", hr);
 
-    hr = IDirect3D2_EnumDevices(d3d, test_enum_devices_caps_callback, ddraw);
+    hr = IDirect3D2_EnumDevices(d3d, test_enum_devices_caps_callback, NULL);
     ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
 
     IDirect3D2_Release(d3d);
@@ -17110,8 +16921,8 @@
 
     IDirect3DViewport2_Release(viewport);
     IDirectDrawSurface_Release(surface);
+    IDirectDraw2_Release(ddraw);
     IDirect3DDevice2_Release(device);
-    IDirectDraw2_Release(ddraw);
     DestroyWindow(window);
 }
 
@@ -17156,291 +16967,10 @@
     DestroyWindow(window);
 }
 
-static void test_yuv_blit(void)
-{
-    DDSURFACEDESC surface_desc = {sizeof(surface_desc)};
-    IDirectDrawSurface *rgb_surface, *yuv_surface;
-    const unsigned int width = 32, height = 32;
-    uint8_t *buf, *chroma_buf, *u_buf, *v_buf;
-    DDBLTFX fx = {.dwSize = sizeof(fx)};
-    unsigned int color, refcount;
-    IDirectDraw2 *ddraw;
-    HWND window;
-    HRESULT hr;
-
-    static const struct
-    {
-        uint8_t y, u, v;
-        uint32_t rgb_full, rgb_reduced;
-    }
-    tests[] =
-    {
-        {0x10, 0x80, 0x80, 0x000000, 0x101010},
-        {0xeb, 0x80, 0x80, 0xffffff, 0xebebeb},
-        {0x51, 0x5a, 0xf0, 0xff0000, 0xee0e0e},
-        {0x91, 0x36, 0x22, 0x00ff01, 0x0dee0e},
-        {0x29, 0xf0, 0x6e, 0x0000ff, 0x100fef},
-        {0x7e, 0x80, 0x80, 0x808080, 0x7e7e7e},
-        {0x00, 0x80, 0x80, 0x000000, 0x000000},
-        {0xff, 0x80, 0x80, 0xffffff, 0xffffff},
-        {0x00, 0x00, 0x00, 0x008800, 0x008800},
-        {0xff, 0x00, 0x00, 0x4aff14, 0x4cff1c},
-        {0x00, 0xff, 0x00, 0x0024ee, 0x0030e1},
-        {0x00, 0x00, 0xff, 0xb80000, 0xb20000},
-        {0xff, 0xff, 0x00, 0x4affff, 0x4cffff},
-        {0xff, 0x00, 0xff, 0xffe114, 0xffd01c},
-        {0x00, 0xff, 0xff, 0xb800ee, 0xb200e1},
-        {0xff, 0xff, 0xff, 0xff7dff, 0xff78ff},
-    };
-
-    static const struct
-    {
-        DWORD fourcc;
-        const char *str;
-    }
-    formats[] =
-    {
-        {MAKEFOURCC('U','Y','V','Y'), "UYVY"},
-        {MAKEFOURCC('Y','U','Y','2'), "YUY2"},
-        {MAKEFOURCC('Y','V','1','2'), "YV12"},
-        {MAKEFOURCC('N','V','1','2'), "NV12"},
-    };
-
-    window = create_window();
-    ddraw = create_ddraw();
-    hr = IDirectDraw2_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    surface_desc.dwFlags = DDSD_WIDTH | DDSD_HEIGHT | DDSD_CAPS | DDSD_PIXELFORMAT;
-    surface_desc.ddsCaps.dwCaps = DDSCAPS_OFFSCREENPLAIN;
-    surface_desc.dwWidth = 640;
-    surface_desc.dwHeight = 480;
-    init_format_b8g8r8x8(&surface_desc.ddpfPixelFormat);
-
-    hr = IDirectDraw2_CreateSurface(ddraw, &surface_desc, &rgb_surface, NULL);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    for (unsigned int fmt = 0; fmt < ARRAY_SIZE(formats); fmt++)
-    {
-        DWORD format = formats[fmt].fourcc;
-
-        winetest_push_context("format %s", formats[fmt].str);
-
-        surface_desc.dwWidth = width;
-        surface_desc.dwHeight = height;
-        surface_desc.ddpfPixelFormat.dwFlags = DDPF_FOURCC;
-        surface_desc.ddpfPixelFormat.dwFourCC = format;
-        hr = IDirectDraw2_CreateSurface(ddraw, &surface_desc, &yuv_surface, NULL);
-        if (hr != S_OK)
-        {
-            skip("Failed to create surface, hr %#lx.\n", hr);
-            winetest_pop_context();
-            continue;
-        }
-
-        for (unsigned int i = 0; i < ARRAY_SIZE(tests); i++)
-        {
-            winetest_push_context("value (%#x,%#x,%#x)", tests[i].y, tests[i].u, tests[i].v);
-
-            hr = IDirectDrawSurface_Lock(yuv_surface, NULL, &surface_desc, DDLOCK_WAIT, NULL);
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-            buf = surface_desc.lpSurface;
-            chroma_buf = buf + surface_desc.lPitch * height;
-            if (format == MAKEFOURCC('Y','V','1','2'))
-            {
-                v_buf = chroma_buf;
-                u_buf = chroma_buf + height / 2 * surface_desc.lPitch / 2;
-            }
-            for (unsigned int y = 0; y < height; y++)
-            {
-                for (unsigned int x = 0; x < width; x += 2)
-                {
-                    uint8_t Y = tests[i].y, U = tests[i].u, V = tests[i].v;
-                    if (x < width / 2 && y < height / 2)
-                        Y = U = V = 0x40;
-
-                    if (format == MAKEFOURCC('U','Y','V','Y'))
-                    {
-                        buf[y * surface_desc.lPitch + 2 * x + 0] = U;
-                        buf[y * surface_desc.lPitch + 2 * x + 1] = Y;
-                        buf[y * surface_desc.lPitch + 2 * x + 2] = V;
-                        buf[y * surface_desc.lPitch + 2 * x + 3] = Y;
-                    }
-                    else if (format == MAKEFOURCC('Y','U','Y','2'))
-                    {
-                        buf[y * surface_desc.lPitch + 2 * x + 0] = Y;
-                        buf[y * surface_desc.lPitch + 2 * x + 1] = U;
-                        buf[y * surface_desc.lPitch + 2 * x + 2] = Y;
-                        buf[y * surface_desc.lPitch + 2 * x + 3] = V;
-                    }
-                    else if (format == MAKEFOURCC('Y','V','1','2'))
-                    {
-                        buf[y * surface_desc.lPitch + x + 0] = Y;
-                        buf[y * surface_desc.lPitch + x + 1] = Y;
-                        u_buf[(y / 2) * (surface_desc.lPitch / 2) + (x / 2)] = U;
-                        v_buf[(y / 2) * (surface_desc.lPitch / 2) + (x / 2)] = V;
-                    }
-                    else if (format == MAKEFOURCC('N','V','1','2'))
-                    {
-                        buf[y * surface_desc.lPitch + x + 0] = Y;
-                        buf[y * surface_desc.lPitch + x + 1] = Y;
-                        chroma_buf[(y / 2) * surface_desc.lPitch + 2 * (x / 2) + 0] = U;
-                        chroma_buf[(y / 2) * surface_desc.lPitch + 2 * (x / 2) + 1] = V;
-                    }
-                }
-            }
-            hr = IDirectDrawSurface_Unlock(yuv_surface, NULL);
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-
-            hr = IDirectDrawSurface_Blt(rgb_surface, NULL, yuv_surface, NULL, DDBLT_WAIT, NULL);
-            if (hr != D3D_OK)
-            {
-                winetest_pop_context();
-                skip("Failed to blit, hr %#lx.\n", hr);
-                break;
-            }
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-
-            hr = IDirectDrawSurface_Lock(rgb_surface, NULL, &surface_desc, DDLOCK_WAIT, NULL);
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-
-            for (unsigned int y = 0; y < 4; y++)
-            {
-                for (unsigned int x = 0; x < 4; x++)
-                {
-                    unsigned int xcoord = (1 + 2 * x) * 640 / 8;
-                    unsigned int ycoord = (1 + 2 * y) * 480 / 8;
-
-                    color = ((uint32_t *)((uint8_t *)surface_desc.lpSurface + ycoord * surface_desc.lPitch))[xcoord];
-                    color &= 0xffffff;
-
-                    if (x < 2 && y < 2)
-                        ok(compare_color(color, 0x008400, 1),
-                                "Got color %#x at (%u, %u).\n", color, xcoord, ycoord);
-                    else
-                        ok(compare_color(color, tests[i].rgb_full, 1)
-                                || compare_color(color, tests[i].rgb_reduced, 1),
-                                "Got color %#x at (%u, %u), expected %#x.\n", color, xcoord, ycoord, tests[i].rgb_full);
-                }
-            }
-            hr = IDirectDrawSurface_Unlock(rgb_surface, NULL);
-            ok(hr == D3D_OK, "Got hr %#lx.\n", hr);
-
-            winetest_pop_context();
-        }
-
-        IDirectDrawSurface_Release(yuv_surface);
-        winetest_pop_context();
-    }
-
-    IDirectDrawSurface_Release(rgb_surface);
-    refcount = IDirectDraw2_Release(ddraw);
-    ok(!refcount, "Device has %u references left.\n", refcount);
-    DestroyWindow(window);
-}
-
-static void test_blit_to_self(void)
-{
-    DDSURFACEDESC surface_desc = {sizeof(surface_desc)};
-    const unsigned int width = 64, height = 64;
-    IDirectDrawSurface *surface;
-    unsigned int refcount;
-    IDirectDraw2 *ddraw;
-    HWND window;
-    HRESULT hr;
-
-    struct
-    {
-        RECT src, dst;
-    }
-    tests[] =
-    {
-        {{16, 16, 24, 24}, {18, 18, 26, 26}},
-        {{16, 16, 24, 24}, {14, 14, 22, 22}},
-        {{16, 16, 24, 24}, {12, 12, 28, 28}},
-        {{16, 16, 24, 24}, {18, 18, 22, 22}},
-    };
-
-    window = create_window();
-    ddraw = create_ddraw();
-    hr = IDirectDraw2_SetCooperativeLevel(ddraw, window, DDSCL_NORMAL);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    surface_desc.dwFlags = DDSD_WIDTH | DDSD_HEIGHT | DDSD_CAPS | DDSD_PIXELFORMAT;
-    surface_desc.ddsCaps.dwCaps = DDSCAPS_OFFSCREENPLAIN | DDSCAPS_SYSTEMMEMORY;
-    surface_desc.dwWidth = width;
-    surface_desc.dwHeight = height;
-    init_format_b8g8r8x8(&surface_desc.ddpfPixelFormat);
-
-    hr = IDirectDraw2_CreateSurface(ddraw, &surface_desc, &surface, NULL);
-    ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-    for (unsigned int i = 0; i < ARRAY_SIZE(tests); ++i)
-    {
-        hr = IDirectDrawSurface_Lock(surface, NULL, &surface_desc, DDLOCK_WAIT, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-        for (unsigned int y = 0; y < height; ++y)
-        {
-            for (unsigned int x = 0; x < width; ++x)
-                ((unsigned int *)surface_desc.lpSurface)[y * width + x] = (y << 8) | x;
-        }
-        hr = IDirectDrawSurface_Unlock(surface, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-        hr = IDirectDrawSurface_Blt(surface, &tests[i].dst, surface, &tests[i].src, DDBLT_WAIT, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-
-        hr = IDirectDrawSurface_Lock(surface, NULL, &surface_desc, DDLOCK_WAIT | DDLOCK_READONLY, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-        for (unsigned int y = 0; y < height; ++y)
-        {
-            for (unsigned int x = 0; x < width; ++x)
-            {
-                unsigned int colour = ((unsigned int *)surface_desc.lpSurface)[y * width + x];
-
-                if (x >= tests[i].dst.left && x < tests[i].dst.right
-                        && y >= tests[i].dst.top && y < tests[i].dst.bottom)
-                {
-                    unsigned int src_x = (x - tests[i].dst.left)
-                            * (tests[i].src.right - tests[i].src.left)
-                            / (tests[i].dst.right - tests[i].dst.left)
-                            + tests[i].src.left;
-                    unsigned int src_y = (y - tests[i].dst.top)
-                            * (tests[i].src.bottom - tests[i].src.top)
-                            / (tests[i].dst.bottom - tests[i].dst.top)
-                            + tests[i].src.top;
-                    if (colour != ((src_y << 8) | src_x))
-                    {
-                        ok(0, "Got colour 0x%08x at (%u, %u), expected 0x%08x.\n",
-                                colour, x, y, ((src_y << 8) | src_x));
-                        goto out;
-                    }
-                }
-                else
-                {
-                    if (colour != ((y << 8) | x))
-                    {
-                        ok(0, "Got colour 0x%08x at (%u, %u).\n", colour, x, y);
-                        goto out;
-                    }
-                }
-            }
-        }
-out:
-        hr = IDirectDrawSurface_Unlock(surface, NULL);
-        ok(hr == S_OK, "Got hr %#lx.\n", hr);
-    }
-
-    IDirectDrawSurface_Release(surface);
-    refcount = IDirectDraw2_Release(ddraw);
-    ok(!refcount, "Device has %u references left.\n", refcount);
-    DestroyWindow(window);
-}
-
 START_TEST(ddraw2)
 {
     DDDEVICEIDENTIFIER identifier;
-    DEVMODEA current_mode;
+    DEVMODEW current_mode;
     IDirectDraw2 *ddraw;
     HMODULE dwmapi;
 
@@ -17462,9 +16992,9 @@
 
     memset(&current_mode, 0, sizeof(current_mode));
     current_mode.dmSize = sizeof(current_mode);
-    ok(EnumDisplaySettingsA(NULL, ENUM_CURRENT_SETTINGS, &current_mode), "Failed to get display mode.\n");
+    ok(EnumDisplaySettingsW(NULL, ENUM_CURRENT_SETTINGS, &current_mode), "Failed to get display mode.\n");
     registry_mode.dmSize = sizeof(registry_mode);
-    ok(EnumDisplaySettingsA(NULL, ENUM_REGISTRY_SETTINGS, &registry_mode), "Failed to get display mode.\n");
+    ok(EnumDisplaySettingsW(NULL, ENUM_REGISTRY_SETTINGS, &registry_mode), "Failed to get display mode.\n");
     if (registry_mode.dmPelsWidth != current_mode.dmPelsWidth
             || registry_mode.dmPelsHeight != current_mode.dmPelsHeight)
     {
@@ -17474,8 +17004,6 @@
 
     if ((dwmapi = LoadLibraryA("dwmapi.dll")))
         pDwmIsCompositionEnabled = (void *)GetProcAddress(dwmapi, "DwmIsCompositionEnabled");
-
-    start_foreground_window_thread();
 
     test_coop_level_create_device_window();
     test_clipper_blt();
@@ -17567,6 +17095,4 @@
     test_multiple_devices();
     test_d3d_state_reset();
     test_sysmem_x_channel();
-    test_yuv_blit();
-    test_blit_to_self();
 }
