diff --git a/dlls/d3dx9_36/tests/surface.c b/dlls/d3dx9_36/tests/surface.c
--- a/dlls/d3dx9_36/tests/surface.c
+++ b/dlls/d3dx9_36/tests/surface.c
@@ -154,10 +154,12 @@
     header->caps = DDSCAPS_TEXTURE;
 }
 
-static void check_dds_pixel_format_image_info(unsigned int line,
+#define check_dds_pixel_format(flags, fourcc, bpp, rmask, gmask, bmask, amask, format) \
+        check_dds_pixel_format_(__LINE__, flags, fourcc, bpp, rmask, gmask, bmask, amask, format)
+static void check_dds_pixel_format_(unsigned int line,
                                     DWORD flags, DWORD fourcc, DWORD bpp,
                                     DWORD rmask, DWORD gmask, DWORD bmask, DWORD amask,
-                                    HRESULT expected_hr, D3DFORMAT expected_format)
+                                    D3DFORMAT expected_format)
 {
     HRESULT hr;
     D3DXIMAGE_INFO info;
@@ -181,9 +183,9 @@
     memset(dds.data, 0, sizeof(dds.data));
 
     hr = D3DXGetImageInfoFromFileInMemory(&dds, sizeof(dds), &info);
-    ok_(__FILE__, line)(hr == expected_hr, "D3DXGetImageInfoFromFileInMemory returned %#lx for pixel format %#x, expected %#lx\n",
-            hr, expected_format, expected_hr);
-    if (SUCCEEDED(hr) && hr == expected_hr)
+    ok_(__FILE__, line)(hr == D3D_OK, "D3DXGetImageInfoFromFileInMemory returned %#lx for pixel format %#x, expected %#lx\n",
+            hr, expected_format, D3D_OK);
+    if (SUCCEEDED(hr))
     {
         ok_(__FILE__, line)(info.Format == expected_format, "D3DXGetImageInfoFromFileInMemory returned format %#x, expected %#x\n",
                 info.Format, expected_format);
@@ -191,7 +193,7 @@
 
     /* Test again with unused fields set. */
     if (flags & DDS_PF_FOURCC)
-        rmask = gmask = bmask = amask = bpp = flags = ~0u;
+        rmask = gmask = bmask = amask = bpp = ~0u;
     else if ((flags & (DDS_PF_INDEXED | DDS_PF_ALPHA)) == (DDS_PF_INDEXED | DDS_PF_ALPHA))
         rmask = gmask = bmask = fourcc = ~0u;
     else if (flags & DDS_PF_INDEXED)
@@ -211,7 +213,6 @@
     else if (flags & DDS_PF_BUMPLUMINANCE)
         fourcc = amask = ~0u;
 
-    dds.header.pixel_format.flags = flags;
     dds.header.pixel_format.fourcc = fourcc;
     dds.header.pixel_format.bpp = bpp;
     dds.header.pixel_format.rmask = rmask;
@@ -219,17 +220,10 @@
     dds.header.pixel_format.bmask = bmask;
     dds.header.pixel_format.amask = amask;
     hr = D3DXGetImageInfoFromFileInMemory(&dds, sizeof(dds), &info);
-    ok_(__FILE__, line)(hr == expected_hr, "Unexpected hr %#lx.\n", hr);
-    if (SUCCEEDED(hr) && hr == expected_hr)
+    ok_(__FILE__, line)(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+    if (SUCCEEDED(hr))
         ok_(__FILE__, line)(info.Format == expected_format, "Unexpected format %#x.\n", info.Format);
 }
-
-#define check_dds_pixel_format(flags, fourcc, bpp, rmask, gmask, bmask, amask, format) \
-        check_dds_pixel_format_image_info(__LINE__, flags, fourcc, bpp, rmask, gmask, bmask, amask, D3D_OK, format)
-
-#define check_dds_pixel_format_unsupported(flags, fourcc, bpp, rmask, gmask, bmask, amask, expected_hr) \
-        check_dds_pixel_format_image_info(__LINE__, flags, fourcc, bpp, rmask, gmask, bmask, amask, expected_hr, \
-                D3DFMT_UNKNOWN)
 
 static void test_dds_header_handling(void)
 {
@@ -500,7 +494,7 @@
 #define IMAGE_RIGHTTOLEFT 0x10
 #define IMAGE_TOPTOBOTTOM 0x20
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 struct tga_header
 {
     uint8_t  id_length;
@@ -523,7 +517,7 @@
     uint32_t developer_directory_offset;
     uint8_t magic[18];
 };
-#pragma pack(pop)
+#include "poppack.h"
 
 static const struct tga_footer default_tga_footer = {
     0, 0,
@@ -967,7 +961,6 @@
     check_dds_pixel_format(DDS_PF_FOURCC, D3DFMT_R32F, 0, 0, 0, 0, 0, D3DFMT_R32F);
     check_dds_pixel_format(DDS_PF_FOURCC, D3DFMT_G32R32F, 0, 0, 0, 0, 0, D3DFMT_G32R32F);
     check_dds_pixel_format(DDS_PF_FOURCC, D3DFMT_A32B32G32R32F, 0, 0, 0, 0, 0, D3DFMT_A32B32G32R32F);
-    check_dds_pixel_format(DDS_PF_FOURCC, D3DFMT_CxV8U8, 0, 0, 0, 0, 0, D3DFMT_CxV8U8);
     check_dds_pixel_format(DDS_PF_RGB, 0, 16, 0xf800, 0x07e0, 0x001f, 0, D3DFMT_R5G6B5);
     check_dds_pixel_format(DDS_PF_RGB | DDS_PF_ALPHA, 0, 16, 0x7c00, 0x03e0, 0x001f, 0x8000, D3DFMT_A1R5G5B5);
     check_dds_pixel_format(DDS_PF_RGB | DDS_PF_ALPHA, 0, 16, 0x0f00, 0x00f0, 0x000f, 0xf000, D3DFMT_A4R4G4B4);
@@ -998,13 +991,6 @@
     check_dds_pixel_format(DDS_PF_INDEXED, 0, 8, 0, 0, 0, 0, D3DFMT_P8);
     check_dds_pixel_format(DDS_PF_INDEXED | DDS_PF_ALPHA, 0, 16, 0, 0, 0, 0xff00, D3DFMT_A8P8);
 
-    /*
-     * ATI{1,2} are unsupported, but some games (Secret World Legends) attempt
-     * to use them on d3dx9.
-     */
-    check_dds_pixel_format_unsupported(DDS_PF_FOURCC, MAKEFOURCC('A','T','I','1'), 0, 0, 0, 0, 0, D3DXERR_INVALIDDATA);
-    check_dds_pixel_format_unsupported(DDS_PF_FOURCC, MAKEFOURCC('A','T','I','2'), 0, 0, 0, 0, 0, D3DXERR_INVALIDDATA);
-
     test_dds_header_handling();
     test_tga_header_handling();
 
@@ -1041,6 +1027,13 @@
 {
    DWORD color = ((DWORD*)lockrect->pBits)[x + y * lockrect->Pitch / 4];
    ok_(__FILE__, line)(color == expected_color, "Got color 0x%08lx, expected 0x%08lx\n", color, expected_color);
+}
+
+#define check_pixel_4bpp_diff(lockrect, x, y, color, max_diff) _check_pixel_4bpp_diff(__LINE__, lockrect, x, y, color, max_diff)
+static inline void _check_pixel_4bpp_diff(unsigned int line, const D3DLOCKED_RECT *lockrect, int x, int y, DWORD expected_color, uint8_t max_diff)
+{
+   DWORD color = ((DWORD*)lockrect->pBits)[x + y * lockrect->Pitch / 4];
+   ok_(__FILE__, line)(compare_color_4bpp(color, expected_color, max_diff), "Got color 0x%08lx, expected 0x%08lx.\n", color, expected_color);
 }
 
 #define check_pixel_8bpp(lockrect, x, y, color) _check_pixel_8bpp(__LINE__, lockrect, x, y, color)
@@ -1152,6 +1145,42 @@
 {
     0x00,0x00,0x00,0x10,0x00,0x00,0x40,0x20,0x00,0x00,0x80,0x30,0xff,0xff,0xff,0x40,
 };
+
+static uint32_t get_bpp_for_d3dformat(D3DFORMAT format)
+{
+    switch (format)
+    {
+    case D3DFMT_A32B32G32R32F:
+        return 16;
+
+    case D3DFMT_A16B16G16R16:
+    case D3DFMT_Q16W16V16U16:
+        return 8;
+
+    case D3DFMT_A8B8G8R8:
+    case D3DFMT_A8R8G8B8:
+    case D3DFMT_V16U16:
+    case D3DFMT_G16R16:
+        return 4;
+
+    case D3DFMT_R8G8B8:
+        return 3;
+
+    case D3DFMT_X1R5G5B5:
+    case D3DFMT_A1R5G5B5:
+    case D3DFMT_V8U8:
+    case D3DFMT_A8P8:
+        return 2;
+
+    case D3DFMT_L8:
+    case D3DFMT_P8:
+        return 1;
+
+    default:
+        assert(0 && "Need to add format to get_bpp_for_d3dformat().");
+        return 0;
+    }
+}
 
 static void test_format_conversion(IDirect3DDevice9 *device)
 {
@@ -1326,7 +1355,7 @@
             {
                 const uint32_t expected_pixel = !(x & 0x01) ? 0xffffffff : 0x00000000;
 
-                check_readback_pixel_4bpp(&surface_rb, x, y, expected_pixel, !expected_pixel);
+                check_readback_pixel_4bpp(&surface_rb, x, y, expected_pixel, FALSE);
             }
         }
         release_surface_readback(&surface_rb);
@@ -1354,9 +1383,8 @@
             for (x = 0; x < 4; ++x)
             {
                 const uint32_t expected_pixel = dxt_pma_decompressed_expected[(y * 4) + x];
-                const BOOL todo = ((expected_pixel >> 24) & 0xff) != 0xff;
-
-                todo_wine_if(todo) check_pixel_4bpp(&lock_rect, x, y, expected_pixel);
+
+                check_pixel_4bpp_diff(&lock_rect, x, y, expected_pixel, 1);
             }
         }
         IDirect3DSurface9_UnlockRect(decomp_surf);
@@ -1993,437 +2021,6 @@
     free(tga);
 }
 
-static const uint8_t r8g8b8_4_4[] =
-{
-    0xff,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0xff,0x00,0xff,0x00,0x00,0xff,
-    0x00,0x00,0x00,0xff,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0xff,0xff,0xff,
-    0x00,0xff,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0xff,0xff,0xff,0x00,0x80,0x80,0x00,
-};
-
-static const uint8_t r8g8b8_4_4_expected[] =
-{
-    0xff,0x00,0x00,0xff,0xff,0x00,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,
-    0xff,0x00,0x00,0xff,0xff,0x00,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,
-    0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0xff,0xff,0x00,0xff,0xff,0xff,0x00,0xff,
-    0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0xff,0xff,0x00,0xff,0x00,0x00,0x00,0x00,
-};
-
-static const uint8_t dxt5_4_4[] =
-{
-    0xff,0x00,0x00,0x00,0x00,0x49,0x92,0x24,0x00,0xf8,0x00,0xf8,0x00,0x00,0x00,0x00,
-};
-
-static const uint8_t dxt5_4_4_expected_no_ck[] =
-{
-    0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,
-    0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,
-    0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,
-    0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,
-};
-
-static const uint8_t dxt5_4_4_expected_ck[] =
-{
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,
-    0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,
-};
-
-static const uint8_t dxt4_4_4[] =
-{
-    0x22,0xcc,0x86,0xc6,0xe6,0x86,0xc6,0xe6,0x10,0x84,0x10,0x84,0x00,0x00,0x00,0x00,
-};
-
-static const uint8_t dxt4_4_4_expected_no_ck[] =
-{
-    0x00,0x00,0x00,0x00,0xff,0xff,0xff,0x22,0xff,0xff,0xff,0x44,0xff,0xff,0xff,0x66,
-    0xf7,0xf3,0xf7,0x88,0xc5,0xc2,0xc5,0xaa,0xa5,0xa2,0xa5,0xcc,0x84,0x82,0x84,0xff,
-    0x00,0x00,0x00,0x00,0xff,0xff,0xff,0x22,0xff,0xff,0xff,0x44,0xff,0xff,0xff,0x66,
-    0xf7,0xf3,0xf7,0x88,0xc5,0xc2,0xc5,0xaa,0xa5,0xa2,0xa5,0xcc,0x84,0x82,0x84,0xff,
-};
-
-static const uint8_t dxt4_4_4_expected_ck[] =
-{
-    0x00,0x00,0x00,0x00,0xff,0xff,0xff,0x22,0x00,0x00,0x00,0x00,0xff,0xff,0xff,0x66,
-    0xf7,0xf3,0xf7,0x88,0xc5,0xc2,0xc5,0xaa,0xa5,0xa2,0xa5,0xcc,0x84,0x82,0x84,0xff,
-    0x00,0x00,0x00,0x00,0xff,0xff,0xff,0x22,0x00,0x00,0x00,0x00,0xff,0xff,0xff,0x66,
-    0xf7,0xf3,0xf7,0x88,0xc5,0xc2,0xc5,0xaa,0xa5,0xa2,0xa5,0xcc,0x84,0x82,0x84,0xff,
-};
-
-static const uint8_t p8_4_4[] =
-{
-    0x10,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xf0,0xf0,0xf0,0xf0,0xf0,0xf0,0xf0,0xf0,
-};
-
-static const uint8_t p8_4_4_expected[] =
-{
-    0x00,0x40,0x00,0x10,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-};
-
-static const uint8_t a8p8_4_4[] =
-{
-    0xff,0x10,0xff,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,0x00,
-    0xf0,0x10,0xf0,0x10,0xf0,0x10,0xf0,0x10,0xf0,0x10,0xf0,0x10,0xf0,0x10,0xf0,0x10,
-};
-
-static const uint8_t a8p8_4_4_expected[] =
-{
-    0x00,0x00,0x00,0x00,0x00,0xc0,0xc0,0x00,0x00,0xc0,0xc0,0x00,0x00,0xc0,0xc0,0x00,
-    0x00,0xc0,0xc0,0x00,0x00,0xc0,0xc0,0x00,0x00,0xc0,0xc0,0x00,0x00,0xc0,0xc0,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-};
-
-static const uint8_t a32r32g32b32_4_4[] =
-{
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-};
-
-static const uint8_t a32r32g32b32_4_4_expected[] =
-{
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,
-    0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,
-};
-
-static const uint8_t r32_4_4[] =
-{
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x80,0x3f,0x00,0x00,0x80,0x3f,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x80,0x3f,0x00,0x00,0x80,0x3f,0x00,0x00,0x80,0x3f,0x00,0x00,0x80,0x3f,
-    0x00,0x00,0x00,0x3f,0x00,0x00,0x00,0x3f,0x00,0x00,0x00,0x3f,0x00,0x00,0x00,0x3f,
-    0x00,0x00,0x00,0x3f,0x00,0x00,0x00,0x3f,0x00,0x00,0x00,0x3f,0x00,0x00,0x00,0x3f,
-};
-
-static const uint8_t r32_4_4_expected[] =
-{
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0xff,0xff,0x80,0xff,0xff,0xff,0x80,0xff,0xff,0xff,0x80,0xff,0xff,0xff,0x80,0xff,
-    0xff,0xff,0x80,0xff,0xff,0xff,0x80,0xff,0xff,0xff,0x80,0xff,0xff,0xff,0x80,0xff,
-};
-
-static const uint8_t q8w8v8u8_4_4[] =
-{
-    0x00,0x10,0x20,0x30,0x40,0x50,0x60,0x70,0x80,0x90,0xa0,0xb0,0xc0,0xd0,0xe0,0xf0,
-    0x04,0x14,0x24,0x34,0x44,0x54,0x64,0x74,0x84,0x94,0xa4,0xb4,0xc4,0xd4,0xe4,0xf4,
-    0x00,0x00,0x28,0x38,0x48,0x58,0x68,0x78,0x88,0x98,0xa8,0xb8,0xc8,0xd8,0xe8,0xf8,
-    0x0c,0x1c,0x2c,0x3c,0x4c,0x5c,0x6c,0x7c,0x8c,0x9c,0xac,0xbc,0xcc,0xdc,0xec,0xfc,
-};
-
-static const uint8_t q8w8v8u8_4_4_expected[] =
-{
-    0xa0,0x90,0x80,0xb0,0xe0,0xd0,0xc0,0xf0,0x1f,0x0f,0x00,0x2f,0x5f,0x4f,0x3f,0x6f,
-    0xa4,0x94,0x84,0xb4,0xe4,0xd4,0xc4,0xf4,0x23,0x13,0x03,0x33,0x63,0x53,0x43,0x73,
-    0x80,0x80,0x80,0x80,0xe8,0xd8,0xc8,0xf8,0x27,0x17,0x07,0x37,0x67,0x57,0x47,0x77,
-    0xac,0x9c,0x8c,0xbc,0xec,0xdc,0xcc,0xfc,0x2b,0x1b,0x0b,0x3b,0x6b,0x5b,0x4b,0x7b,
-};
-
-static const uint8_t a4r4g4b4_4_4[] =
-{
-    0xff,0x00,0x0f,0x00,0x00,0xf0,0x00,0x0f,0xf0,0x00,0xf0,0x00,0xf0,0x00,0xf0,0x00,
-    0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,0x80,
-};
-
-static const uint8_t a4r4g4b4_4_4_expected[] =
-{
-    0xff,0xff,0x00,0x00,0xff,0x00,0x00,0x00,0x00,0x00,0x00,0xff,0x00,0x00,0xff,0x00,
-    0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-};
-
-static const uint8_t a4r4g4b4_4_4_expected2[] =
-{
-    0x00,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0x00,0x00,0x00,0xff,0x00,0x00,0xff,0x00,
-    0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x00,
-    0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,
-    0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,0x00,0x88,
-};
-
-static const uint8_t a4l4_4_4[] =
-{
-    0x00,0x11,0x22,0x33,0x44,0x55,0x66,0x77,0x88,0x99,0xaa,0xbb,0xcc,0xdd,0xee,0xff,
-};
-
-static const uint8_t a4l4_4_4_expected[] =
-{
-    0x00,0x00,0x00,0x00,0x11,0x11,0x11,0x11,0x22,0x22,0x22,0x22,0x33,0x33,0x33,0x33,
-    0x44,0x44,0x44,0x44,0x55,0x55,0x55,0x55,0x66,0x66,0x66,0x66,0x77,0x77,0x77,0x77,
-    0x00,0x00,0x00,0x00,0x99,0x99,0x99,0x99,0xaa,0xaa,0xaa,0xaa,0xbb,0xbb,0xbb,0xbb,
-    0xcc,0xcc,0xcc,0xcc,0xdd,0xdd,0xdd,0xdd,0xee,0xee,0xee,0xee,0xff,0xff,0xff,0xff,
-};
-
-static const uint8_t a4l4_4_4_expected2[] =
-{
-    0x00,0x00,0x00,0x00,0x11,0x11,0x11,0x11,0x22,0x22,0x22,0x22,0x33,0x33,0x33,0x33,
-    0x44,0x44,0x44,0x44,0x55,0x55,0x55,0x55,0x66,0x66,0x66,0x66,0x77,0x77,0x77,0x77,
-    0x88,0x88,0x88,0x88,0x99,0x99,0x99,0x99,0xaa,0xaa,0xaa,0xaa,0xbb,0xbb,0xbb,0xbb,
-    0xcc,0xcc,0xcc,0xcc,0xdd,0xdd,0xdd,0xdd,0xee,0xee,0xee,0xee,0xff,0xff,0xff,0xff,
-};
-
-static const uint8_t a1r5g5b5_4_4[] =
-{
-    0x00,0x00,0x42,0x08,0x84,0x10,0xc6,0x18,0x08,0x21,0x4a,0x29,0x8c,0x31,0xce,0x39,
-    0x10,0xc2,0x52,0xca,0x94,0xd2,0xd6,0xda,0x18,0xe3,0x5a,0xeb,0x9c,0xf3,0xde,0xfb,
-};
-
-static const uint8_t a1r5g5b5_4_4_expected[] =
-{
-    0x00,0x00,0x00,0x00,0x10,0x10,0x10,0x00,0x21,0x21,0x21,0x00,0x31,0x31,0x31,0x00,
-    0x42,0x42,0x42,0x00,0x52,0x52,0x52,0x00,0x63,0x63,0x63,0x00,0x73,0x73,0x73,0x00,
-    0x84,0x84,0x84,0xff,0x94,0x94,0x94,0xff,0x00,0x00,0x00,0x00,0xb5,0xb5,0xb5,0xff,
-    0xc5,0xc5,0xc5,0xff,0xd6,0xd6,0xd6,0xff,0xe6,0xe6,0xe6,0xff,0xf7,0xf7,0xf7,0xff,
-};
-
-static const uint8_t r5g6b5_4_4[] =
-{
-    0x00,0x00,0x82,0x10,0x04,0x21,0x86,0x31,0x08,0x42,0x8a,0x52,0x0c,0x63,0x8e,0x73,
-    0x10,0x84,0x92,0x94,0x14,0xa5,0x96,0xb5,0x18,0xc6,0x9a,0xd6,0x1c,0xe7,0x9e,0xf7,
-};
-
-static const uint8_t r5g6b5_4_4_expected[] =
-{
-    0x00,0x00,0x00,0xff,0x10,0x10,0x10,0xff,0x21,0x20,0x21,0xff,0x31,0x31,0x31,0xff,
-    0x42,0x41,0x42,0xff,0x52,0x51,0x52,0xff,0x63,0x61,0x63,0xff,0x73,0x71,0x73,0xff,
-    0x84,0x82,0x84,0xff,0x94,0x92,0x94,0xff,0x00,0x00,0x00,0x00,0xb5,0xb2,0xb5,0xff,
-    0xc5,0xc2,0xc5,0xff,0xd6,0xd2,0xd6,0xff,0xe6,0xe3,0xe6,0xff,0xf7,0xf3,0xf7,0xff,
-};
-
-static const uint8_t r3g3b2_4_4[] =
-{
-    0x00,0x00,0x24,0x24,0x49,0x49,0x6d,0x6d,0x92,0x92,0xb6,0xb6,0xdb,0xdb,0xff,0xff,
-};
-
-static const uint8_t r3g3b2_4_4_expected[] =
-{
-    0x00,0x00,0x00,0xff,0x00,0x00,0x00,0xff,0x00,0x24,0x24,0xff,0x00,0x24,0x24,0xff,
-    0x55,0x49,0x49,0xff,0x55,0x49,0x49,0xff,0x55,0x6d,0x6d,0xff,0x55,0x6d,0x6d,0xff,
-    0xaa,0x92,0x92,0xff,0xaa,0x92,0x92,0xff,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,
-    0xff,0xdb,0xdb,0xff,0xff,0xdb,0xdb,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,0xff,
-};
-
-static BOOL is_dxt_d3dformat(D3DFORMAT fmt);
-static void test_color_key(void)
-{
-    static const uint8_t a8r8g8b8_4_4[] =
-    {
-        0x00,0xff,0x00,0x80,0x00,0xff,0x00,0x80,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,
-        0x80,0x00,0x00,0xff,0xff,0x00,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,
-        0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0xff,0xff,0x00,0xff,0xff,0xff,0x00,0xff,
-        0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0xff,0xff,0x00,0xff,0xff,0xff,0x00,0xff,
-    };
-    static const uint8_t a8r8g8b8_4_4_expected[] =
-    {
-        0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,
-        0x80,0x00,0x00,0xff,0xff,0x00,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,0x00,0xff,
-        0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0xff,0xff,0x00,0xff,0xff,0xff,0x00,0xff,
-        0x00,0x00,0xff,0xff,0x00,0x00,0xff,0xff,0xff,0xff,0x00,0xff,0xff,0xff,0x00,0xff,
-    };
-
-    static const struct
-    {
-        D3DFORMAT src_format;
-        const void *src_data;
-
-        const void *expected_dst_data;
-        D3DCOLOR color_key;
-        const void *expected_dst_data_32;
-        uint8_t max_diff;
-        BOOL todo;
-    } tests[] =
-    {
-        /* Color key with alpha channel unset. */
-        {
-            D3DFMT_R8G8B8, r8g8b8_4_4, r8g8b8_4_4_expected, 0x00008080,
-        },
-        /* Same color key as before except the alpha channel is set. */
-        {
-            D3DFMT_R8G8B8, r8g8b8_4_4, r8g8b8_4_4_expected, 0xff008080,
-        },
-        /* Color key on a palette. */
-        {
-            D3DFMT_P8, p8_4_4, p8_4_4_expected, 0xf0c0c000,
-        },
-        {
-            D3DFMT_A8P8, a8p8_4_4, a8p8_4_4_expected, 0x10c0c000,
-        },
-        {
-            D3DFMT_A32B32G32R32F, a32r32g32b32_4_4, a32r32g32b32_4_4_expected, 0xffff0000,
-        },
-        /* 5. */
-        /*
-         * Only channels that exist in the source format matter for the color
-         * key, other channels can be set to any value and are ignored.
-         */
-        {
-            D3DFMT_R32F, r32_4_4, r32_4_4_expected, 0xffff3080,
-        },
-        /*
-         * Both 0x7f and 0x80 channel values in the color key map to 0x00 in
-         * 8-bit SNORM.
-         */
-        {
-            D3DFMT_Q8W8V8U8, q8w8v8u8_4_4, q8w8v8u8_4_4_expected, 0xb87f80a8,
-        },
-        /* Alpha channel factors into the color key check. */
-        {
-            D3DFMT_A8R8G8B8, a8r8g8b8_4_4, a8r8g8b8_4_4_expected, 0x8000ff00,
-        },
-        {
-            D3DFMT_A8R8G8B8, a8r8g8b8_4_4, a8r8g8b8_4_4, 0x0000ff00,
-        },
-        /*
-         * 0x80-0x90 color key channel values all match 0x8 in the source
-         * format.
-         */
-        {
-            D3DFMT_A4R4G4B4, a4r4g4b4_4_4, a4r4g4b4_4_4_expected, 0x80009000,
-        },
-        /* 10. */
-        /* 0xf7-0xff match 0xf in the source format. */
-        {
-            D3DFMT_A4R4G4B4, a4r4g4b4_4_4, a4r4g4b4_4_4_expected2, 0x0000f7ff,
-        },
-        {
-            D3DFMT_A4L4, a4l4_4_4, a4l4_4_4_expected, 0x88818283,
-        },
-        {
-            D3DFMT_A4L4, a4l4_4_4, a4l4_4_4_expected2, 0x88818200,
-        },
-        {
-            D3DFMT_A1R5G5B5, a1r5g5b5_4_4, a1r5g5b5_4_4_expected, 0x80a1a8a4, .max_diff = 1
-        },
-        {
-            D3DFMT_R5G6B5, r5g6b5_4_4, r5g6b5_4_4_expected, 0x81a1a0a8, .max_diff = 1
-        },
-        /* 15. */
-        {
-            D3DFMT_R3G3B2, r3g3b2_4_4, r3g3b2_4_4_expected, 0x81a4c8bf,
-        },
-        /*
-         * Test color key handling for compressed formats. On 64-bit the color key
-         * is used, but on 32-bit it is ignored.
-         */
-        {
-            D3DFMT_DXT5, dxt5_4_4, dxt5_4_4_expected_ck, 0xffff0000, dxt5_4_4_expected_no_ck,
-        },
-        /*
-         * Test premultiplied alpha handling with a color key - color key applies
-         * to the value after the premultiplied alpha transformation is undone.
-         */
-        {
-            D3DFMT_DXT4, dxt4_4_4, dxt4_4_4_expected_ck, 0x44ffffff, dxt4_4_4_expected_no_ck,
-            .todo = TRUE
-        },
-        /*
-         * Color key also has the same range of values as R5G6B5 for the color
-         * channels.
-         */
-        {
-            D3DFMT_DXT4, dxt4_4_4, dxt4_4_4_expected_ck, 0x44fbfdfb, dxt4_4_4_expected_no_ck,
-            .todo = TRUE
-        },
-        /* Values out of range for R5G6B5, won't be color keyed. */
-        {
-            D3DFMT_DXT4, dxt4_4_4, dxt4_4_4_expected_no_ck, 0x44fafcfa,
-            .todo = TRUE
-        },
-    };
-    unsigned int i, x, y, mismatch_count;
-    PALETTEENTRY tmp_palette[256];
-    const uint8_t *expected_dst;
-    IDirect3DDevice9 *device;
-    D3DLOCKED_RECT lock_rect;
-    IDirect3DSurface9 *surf;
-    uint32_t src_pitch;
-    RECT rect;
-    HRESULT hr;
-    HWND hwnd;
-
-    if (!(device = create_device(&hwnd)))
-        return;
-
-    memcpy(tmp_palette, test_palette, sizeof(test_palette));
-    /* Set palette entries 0xf1-0xff to the same value. */
-    for (i = 0; i < 0xf; ++i)
-        tmp_palette[0xf1 + i] = tmp_palette[0xf0];
-
-    SetRect(&rect, 0, 0, 4, 4);
-    hr = IDirect3DDevice9_CreateOffscreenPlainSurface(device, 4, 4, D3DFMT_A8R8G8B8, D3DPOOL_SCRATCH, &surf, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    for (i = 0; i < ARRAY_SIZE(tests); ++i)
-    {
-        winetest_push_context("Test %u", i);
-
-        mismatch_count = 0;
-        src_pitch = get_bpp_for_d3dformat(tests[i].src_format) * 4;
-        hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, tests[i].src_data, tests[i].src_format,
-                src_pitch, tmp_palette, &rect, D3DX_FILTER_NONE, tests[i].color_key);
-        ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-        if (sizeof(void *) == 4 && tests[i].expected_dst_data_32)
-            expected_dst = tests[i].expected_dst_data_32;
-        else
-            expected_dst = tests[i].expected_dst_data;
-
-        IDirect3DSurface9_LockRect(surf, &lock_rect, NULL, D3DLOCK_READONLY);
-        for (y = 0; y < 4; ++y)
-        {
-            const uint8_t *dst_expected_row = ((const uint8_t *)expected_dst) + (sizeof(uint32_t) * 4 * y);
-            const uint8_t *dst_row = ((const uint8_t *)lock_rect.pBits) + (lock_rect.Pitch * y);
-
-            for (x = 0; x < 4; ++x)
-            {
-                const uint32_t dst_expected_color = ((const uint32_t *)dst_expected_row)[x];
-                const uint32_t dst_color = ((const uint32_t *)dst_row)[x];
-
-                if (!compare_color_4bpp(dst_color, dst_expected_color, tests[i].max_diff))
-                    mismatch_count++;
-            }
-        }
-        IDirect3DSurface9_UnlockRect(surf);
-
-        todo_wine_if(tests[i].todo) ok(!mismatch_count, "Unexpected number of mismatched pixels %u.\n", mismatch_count);
-        winetest_pop_context();
-    }
-
-    /*
-     * Trying to use D3DX_FILTER_SRGB_IN with a color key results in
-     * STATUS_ACCESS_VIOLATION on native.
-     */
-    if (0)
-    {
-        const uint32_t src_pitch = get_bpp_for_d3dformat(tests[0].src_format) * 4;
-
-        hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, tests[0].src_data, tests[0].src_format,
-                src_pitch, NULL, &rect, D3DX_FILTER_NONE | D3DX_FILTER_SRGB_IN,
-                tests[0].color_key);
-        ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    }
-
-    check_release((IUnknown *)surf, 0);
-    check_release((IUnknown *)device, 0);
-    DestroyWindow(hwnd);
-}
-
 static void test_D3DXLoadSurface(IDirect3DDevice9 *device)
 {
     HRESULT hr;
@@ -2432,7 +2029,6 @@
     IDirect3DSurface9 *surf, *newsurf;
     RECT rect, destrect;
     D3DLOCKED_RECT lockrect;
-    UINT row_pitch;
     static const WORD pixdata_a8r3g3b2[] = { 0x57df, 0x98fc, 0xacdd, 0xc891 };
     static const WORD pixdata_a1r5g5b5[] = { 0x46b5, 0x99c8, 0x06a2, 0x9431 };
     static const WORD pixdata_r5g6b5[] = { 0x9ef6, 0x658d, 0x0aee, 0x42ee };
@@ -2447,10 +2043,8 @@
     static const uint32_t pixdata_q8w8v8u8[] = { 0x30201000, 0x7f605040, 0xb0a08180, 0xffe0d0c0 };
     static const float pixdata_a32b32g32r32f[] = {  0.0f,  0.1f,  NAN,  INFINITY,  1.0f,  1.1f,  1.2f,  1.3f,
                                                    -0.1f, -0.2f, -NAN, -INFINITY, -1.0f, -1.1f, -1.2f, -1.3f };
-    static const uint16_t pixdata_cxv8u8[] = { 0x7f7f, 0x8181, 0x0000, 0x006d };
     static const uint16_t pixdata_v8u8[] = { 0x3000, 0x7f40, 0x8180, 0xffc0 };
     BYTE buffer[4 * 8 * 4];
-    D3DXIMAGE_INFO info;
     uint32_t i;
 
     hr = create_file("testdummy.bmp", noimage, sizeof(noimage));  /* invalid image */
@@ -2534,45 +2128,44 @@
 
     /* D3DXLoadSurfaceFromMemory */
     SetRect(&rect, 0, 0, 2, 2);
-    row_pitch = 8; /* 2 * 4 bytes */
-
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata, D3DFMT_A8R8G8B8, row_pitch, NULL, &rect, D3DX_FILTER_NONE, 0);
+
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, &rect, D3DX_FILTER_NONE, 0);
     ok(hr == D3D_OK, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
 
     hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata, D3DFMT_A8R8G8B8, 0, NULL, &rect, D3DX_FILTER_NONE, 0);
     ok(hr == D3D_OK, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
 
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, NULL, D3DFMT_A8R8G8B8, row_pitch, NULL, &rect, D3DX_DEFAULT, 0);
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, NULL, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, &rect, D3DX_DEFAULT, 0);
     ok(hr == D3DERR_INVALIDCALL, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3DERR_INVALIDCALL);
 
-    hr = D3DXLoadSurfaceFromMemory(NULL, NULL, NULL, pixdata, D3DFMT_A8R8G8B8, row_pitch, NULL, &rect, D3DX_DEFAULT, 0);
+    hr = D3DXLoadSurfaceFromMemory(NULL, NULL, NULL, pixdata, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, &rect, D3DX_DEFAULT, 0);
     ok(hr == D3DERR_INVALIDCALL, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3DERR_INVALIDCALL);
 
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata, D3DFMT_A8R8G8B8, row_pitch, NULL, NULL, D3DX_DEFAULT, 0);
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, NULL, D3DX_DEFAULT, 0);
     ok(hr == D3DERR_INVALIDCALL, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3DERR_INVALIDCALL);
 
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata, D3DFMT_UNKNOWN, row_pitch, NULL, &rect, D3DX_DEFAULT, 0);
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata, D3DFMT_UNKNOWN, sizeof(pixdata), NULL, &rect, D3DX_DEFAULT, 0);
     ok(hr == E_FAIL, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, E_FAIL);
 
     SetRect(&destrect, -1, -1, 1, 1); /* destination rect is partially outside texture boundaries */
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, row_pitch, NULL, &rect, D3DX_FILTER_NONE, 0);
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, &rect, D3DX_FILTER_NONE, 0);
     ok(hr == D3DERR_INVALIDCALL, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3DERR_INVALIDCALL);
 
     SetRect(&destrect, 255, 255, 257, 257); /* destination rect is partially outside texture boundaries */
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, row_pitch, NULL, &rect, D3DX_FILTER_NONE, 0);
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, &rect, D3DX_FILTER_NONE, 0);
     ok(hr == D3DERR_INVALIDCALL, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3DERR_INVALIDCALL);
 
     SetRect(&destrect, 1, 1, 0, 0); /* left > right, top > bottom */
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, row_pitch, NULL, &rect, D3DX_FILTER_NONE, 0);
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, &rect, D3DX_FILTER_NONE, 0);
     ok(hr == D3DERR_INVALIDCALL, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3DERR_INVALIDCALL);
 
     SetRect(&destrect, 1, 2, 1, 2); /* left = right, top = bottom */
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, row_pitch, NULL, &rect, D3DX_FILTER_NONE, 0);
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, &rect, D3DX_FILTER_NONE, 0);
     /* fails when debug version of d3d9 is used */
     ok(hr == D3D_OK || broken(hr == D3DERR_INVALIDCALL), "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3D_OK);
 
     SetRect(&destrect, 257, 257, 257, 257); /* left = right, top = bottom, but invalid values */
-    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, row_pitch, NULL, &rect, D3DX_FILTER_NONE, 0);
+    hr = D3DXLoadSurfaceFromMemory(surf, NULL, &destrect, pixdata, D3DFMT_A8R8G8B8, sizeof(pixdata), NULL, &rect, D3DX_FILTER_NONE, 0);
     ok(hr == D3DERR_INVALIDCALL, "D3DXLoadSurfaceFromMemory returned %#lx, expected %#lx\n", hr, D3DERR_INVALIDCALL);
 
 
@@ -2883,19 +2476,6 @@
         hr = IDirect3DSurface9_UnlockRect(surf);
         ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
 
-        /* D3DFMT_CxV8U8. */
-        hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata_cxv8u8, D3DFMT_CxV8U8, 4, NULL, &rect,
-                D3DX_FILTER_NONE, 0);
-        ok(hr == D3D_OK, "Got unexpected hr %#lx.\n", hr);
-        hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
-        ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
-        check_pixel_4bpp(&lockrect, 0, 0, 0xffffff80);
-        check_pixel_4bpp(&lockrect, 1, 0, 0xff000080);
-        check_pixel_4bpp(&lockrect, 0, 1, 0xff8080ff);
-        check_pixel_4bpp(&lockrect, 1, 1, 0xffed80c1);
-        hr = IDirect3DSurface9_UnlockRect(surf);
-        ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
-
         hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata_a32b32g32r32f, D3DFMT_A32B32G32R32F, 32, NULL, &rect,
                 D3DX_FILTER_NONE, 0);
         ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
@@ -2975,6 +2555,50 @@
         hr = IDirect3DSurface9_UnlockRect(surf);
         ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
 
+        /* Test SRGB filter flags. */
+        SetRect(&rect, 0, 0, 2, 2);
+        hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata_a8r8g8b8, D3DFMT_A8R8G8B8, 8, NULL, &rect,
+                D3DX_FILTER_NONE | D3DX_FILTER_SRGB_IN, 0);
+        ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+        /* SRGB in, no SRGB out. */
+        hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
+        ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
+        check_pixel_4bpp_diff(&lockrect, 0, 0, 0x00010306, 2);
+        check_pixel_4bpp_diff(&lockrect, 1, 0, 0x40141e2a, 2);
+        check_pixel_4bpp_diff(&lockrect, 0, 1, 0x80495b71, 2);
+        check_pixel_4bpp_diff(&lockrect, 1, 1, 0xc0a3c0ff, 2);
+        hr = IDirect3DSurface9_UnlockRect(surf);
+        ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
+
+        hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata_a8r8g8b8, D3DFMT_A8R8G8B8, 8, NULL, &rect,
+                D3DX_FILTER_NONE | D3DX_FILTER_SRGB_OUT, 0);
+        ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+        /* No SRGB in, SRGB out. */
+        hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
+        ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
+        check_pixel_4bpp_diff(&lockrect, 0, 0, 0x00486377, 2);
+        check_pixel_4bpp_diff(&lockrect, 1, 0, 0x4097a4af, 2);
+        check_pixel_4bpp_diff(&lockrect, 0, 1, 0x80c5ced7, 2);
+        check_pixel_4bpp_diff(&lockrect, 1, 1, 0xc0e8f0ff, 2);
+        hr = IDirect3DSurface9_UnlockRect(surf);
+        ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
+
+        hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata_a8r8g8b8, D3DFMT_A8R8G8B8, 8, NULL, &rect,
+                D3DX_FILTER_NONE | D3DX_FILTER_SRGB_IN | D3DX_FILTER_SRGB_OUT, 0);
+        ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+        /* SRGB in, SRGB out, no changes. */
+        hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
+        ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
+        check_pixel_4bpp_diff(&lockrect, 0, 0, pixdata_a8r8g8b8[0], 0);
+        check_pixel_4bpp_diff(&lockrect, 1, 0, pixdata_a8r8g8b8[1], 0);
+        check_pixel_4bpp_diff(&lockrect, 0, 1, pixdata_a8r8g8b8[2], 0);
+        check_pixel_4bpp_diff(&lockrect, 1, 1, pixdata_a8r8g8b8[3], 0);
+        hr = IDirect3DSurface9_UnlockRect(surf);
+        ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
+
         check_release((IUnknown*)surf, 0);
     }
 
@@ -3500,21 +3124,6 @@
         hr = IDirect3DSurface9_UnlockRect(surf);
         ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
 
-        /* D3DFMT_CxV8U8. */
-        hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata_cxv8u8, D3DFMT_CxV8U8, 4, NULL, &rect,
-                D3DX_FILTER_NONE, 0);
-        ok(hr == D3D_OK, "Got unexpected hr %#lx.\n", hr);
-
-        /* The calculated Cx value goes into the blue channel. */
-        hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
-        ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
-        check_pixel_float4(&lockrect, 0, 0,  1.0f,             1.0f, 0.0f,            1.0f, 0, FALSE);
-        check_pixel_float4(&lockrect, 1, 0, -1.0f,            -1.0f, 0.0f,            1.0f, 0, FALSE);
-        check_pixel_float4(&lockrect, 0, 1,  0.0f,             0.0f, 1.0f,            1.0f, 0, FALSE);
-        check_pixel_float4(&lockrect, 1, 1,  8.58267725e-001,  0.0f, 5.13202250e-001, 1.0f, 0, FALSE);
-        hr = IDirect3DSurface9_UnlockRect(surf);
-        ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
-
         check_release((IUnknown*)surf, 0);
     }
 
@@ -3630,20 +3239,6 @@
         hr = IDirect3DSurface9_UnlockRect(surf);
         ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
 
-        /* D3DFMT_CxV8U8. */
-        hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, pixdata_cxv8u8, D3DFMT_CxV8U8, 4, NULL, &rect,
-                D3DX_FILTER_NONE, 0);
-        ok(hr == D3D_OK, "Got unexpected hr %#lx.\n", hr);
-
-        hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
-        ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
-        check_pixel_4bpp(&lockrect, 0, 0, 0x00ff7f7f);
-        check_pixel_4bpp(&lockrect, 1, 0, 0x00ff8282);
-        check_pixel_4bpp(&lockrect, 0, 1, 0x00ff0000);
-        check_pixel_4bpp(&lockrect, 1, 1, 0x00ff006d);
-        hr = IDirect3DSurface9_UnlockRect(surf);
-        ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
-
         check_release((IUnknown*)surf, 1);
         check_release((IUnknown*)tex, 0);
     }
@@ -3732,57 +3327,6 @@
     ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
     check_release((IUnknown *)surf, 0);
 
-    hr = IDirect3DDevice9_CreateOffscreenPlainSurface(device, 2, 2, D3DFMT_A8R8G8B8, D3DPOOL_SCRATCH, &surf, NULL);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-    /* Indexed PNG. */
-    SetRect(&rect, 0, 0, 2, 2);
-    hr = D3DXLoadSurfaceFromFileInMemory(surf, NULL, NULL, png_2_2_1bpp_indexed, sizeof(png_2_2_1bpp_indexed), &rect,
-            D3DX_FILTER_NONE, 0, &info);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    check_image_info(&info, 2, 2, 1, 1, D3DFMT_P8, D3DRTYPE_TEXTURE, D3DXIFF_PNG, FALSE);
-
-    hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
-    ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
-    check_pixel_4bpp(&lockrect, 0, 0, 0xffffffff);
-    check_pixel_4bpp(&lockrect, 1, 0, 0x00000000);
-    check_pixel_4bpp(&lockrect, 0, 1, 0x00000000);
-    check_pixel_4bpp(&lockrect, 1, 1, 0xffffffff);
-    hr = IDirect3DSurface9_UnlockRect(surf);
-    ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
-
-    /* 2bpp. */
-    hr = D3DXLoadSurfaceFromFileInMemory(surf, NULL, NULL, png_2_2_2bpp_indexed, sizeof(png_2_2_2bpp_indexed), &rect,
-            D3DX_FILTER_NONE, 0, &info);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    check_image_info(&info, 2, 2, 1, 1, D3DFMT_P8, D3DRTYPE_TEXTURE, D3DXIFF_PNG, FALSE);
-
-    hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
-    ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
-    check_pixel_4bpp(&lockrect, 0, 0, 0xffff0000);
-    check_pixel_4bpp(&lockrect, 1, 0, 0xff00ff00);
-    check_pixel_4bpp(&lockrect, 0, 1, 0xff0000ff);
-    check_pixel_4bpp(&lockrect, 1, 1, 0xffffffff);
-    hr = IDirect3DSurface9_UnlockRect(surf);
-    ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
-
-    /* 4bpp. */
-    hr = D3DXLoadSurfaceFromFileInMemory(surf, NULL, NULL, png_2_2_4bpp_indexed, sizeof(png_2_2_4bpp_indexed), &rect,
-            D3DX_FILTER_NONE, 0, &info);
-    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-    check_image_info(&info, 2, 2, 1, 1, D3DFMT_P8, D3DRTYPE_TEXTURE, D3DXIFF_PNG, FALSE);
-
-    hr = IDirect3DSurface9_LockRect(surf, &lockrect, NULL, D3DLOCK_READONLY);
-    ok(hr == D3D_OK, "Failed to lock surface, hr %#lx.\n", hr);
-    check_pixel_4bpp(&lockrect, 0, 0, 0xf0000000);
-    check_pixel_4bpp(&lockrect, 1, 0, 0xf0444444);
-    check_pixel_4bpp(&lockrect, 0, 1, 0xf0888888);
-    check_pixel_4bpp(&lockrect, 1, 1, 0xf0cccccc);
-    hr = IDirect3DSurface9_UnlockRect(surf);
-    ok(hr == D3D_OK, "Failed to unlock surface, hr %#lx.\n", hr);
-
-    check_release((IUnknown *)surf, 0);
-
     test_format_conversion(device);
     test_dxt_premultiplied_alpha(device);
     test_load_surface_from_tga(device);
@@ -4073,12 +3617,6 @@
             DDS_FILE_HEADER_SIZE + (4 * 2 * 4)
           }
         },
-        { D3DFMT_CxV8U8, 4, 4, NULL,
-          { D3D_OK, { 32, DDS_PF_FOURCC, D3DFMT_CxV8U8, 0, 0, 0, 0, 0 },
-            DDSD_CAPS | DDSD_HEIGHT | DDSD_WIDTH | DDSD_PIXELFORMAT, 4, 4, 0, 0, 0, DDSCAPS_TEXTURE, 0,
-            DDS_FILE_HEADER_SIZE + (4 * 2 * 4)
-          }
-        },
     };
     struct
     {
@@ -4130,27 +3668,14 @@
 
 static void test_save_surface_iffs(IDirect3DDevice9 *device)
 {
-    static const struct
-    {
-        D3DXIMAGE_FILEFORMAT format;
-        const char *name;
-    }
-    test_iffs[] =
-    {
-        { D3DXIFF_BMP, "D3DXIFF_BMP" },
-        { D3DXIFF_JPG, "D3DXIFF_JPG" },
-        { D3DXIFF_TGA, "D3DXIFF_TGA" },
-        { D3DXIFF_PNG, "D3DXIFF_PNG" },
-        { D3DXIFF_PPM, "D3DXIFF_PPM" },
-        { D3DXIFF_DIB, "D3DXIFF_DIB" },
-        { D3DXIFF_HDR, "D3DXIFF_HDR" },
-        { D3DXIFF_PFM, "D3DXIFF_PFM" },
-    };
+    static const D3DXIMAGE_FILEFORMAT test_iff[] = { D3DXIFF_BMP, D3DXIFF_JPG, D3DXIFF_TGA, D3DXIFF_PNG,
+                                                     D3DXIFF_PPM, D3DXIFF_DIB, D3DXIFF_HDR, D3DXIFF_PFM };
+    static const char *test_iff_str[] = { "D3DXIFF_BMP", "D3DXIFF_JPG", "D3DXIFF_TGA", "D3DXIFF_PNG",
+                                          "D3DXIFF_PPM", "D3DXIFF_DIB", "D3DXIFF_HDR", "D3DXIFF_PFM" };
     static const struct
     {
         D3DFORMAT format;
         const PALETTEENTRY *palette;
-        uint8_t init_pixel_value;
         struct
         {
             HRESULT hr;
@@ -4159,10 +3684,9 @@
             BOOL todo_format;
         }
         iff_expected[8];
-    }
-    iff_tests[] =
-    {
-        { D3DFMT_P8, test_palette, 0x00,
+        uint8_t init_pixel_value;
+    } iff_tests[] = {
+        { D3DFMT_P8, test_palette,
           { { D3D_OK, D3DFMT_P8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4173,7 +3697,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A8P8, test_palette, 0x00,
+        { D3DFMT_A8P8, test_palette,
           { { D3D_OK, D3DFMT_P8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4184,7 +3708,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_P8, NULL, 0x00,
+        { D3DFMT_P8, NULL,
           { { D3D_OK, D3DFMT_A8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A8R8G8B8, .todo_hr = TRUE },
@@ -4195,7 +3719,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A8P8, NULL, 0x00,
+        { D3DFMT_A8P8, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A8R8G8B8, .todo_hr = TRUE },
@@ -4206,7 +3730,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A8P8, NULL, 0xff,
+        { D3DFMT_A8P8, NULL,
           { { D3D_OK, D3DFMT_A8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A8R8G8B8, .todo_hr = TRUE },
@@ -4215,24 +3739,24 @@
             { D3D_OK, D3DFMT_A8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
+          }, 0xff
+        },
+        { D3DFMT_R8G8B8, NULL,
+          { { D3D_OK, D3DFMT_R8G8B8, },
+            { D3D_OK, D3DFMT_X8R8G8B8 },
+            { D3D_OK, D3DFMT_R8G8B8 },
+            { D3D_OK, D3DFMT_X8R8G8B8 },
+            { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
+            { D3D_OK, D3DFMT_R8G8B8, },
+            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
+            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_R8G8B8, NULL, 0x00,
-          { { D3D_OK, D3DFMT_R8G8B8, },
-            { D3D_OK, D3DFMT_X8R8G8B8 },
-            { D3D_OK, D3DFMT_R8G8B8 },
-            { D3D_OK, D3DFMT_X8R8G8B8 },
-            { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_R8G8B8, },
-            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-          },
-        },
         /*
-         * For BMP/DIB, these encode as D3DFMT_X8R8G8B8. If there's at least
-         * one pixel with a non-zero alpha channel, it reports as D3DFMT_A8R8G8B8.
+         * For BMP/DIB, these encode as D3DFMT_X8R8G8B8. If there's a single
+         * pixel with a non-zero alpha channel, it reports as D3DFMT_A8R8G8B8.
          */
-        { D3DFMT_A8R8G8B8, NULL, 0x00,
+        { D3DFMT_A8R8G8B8, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4243,7 +3767,11 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A8R8G8B8, NULL, 0xff,
+        /*
+         * Same as decoding for BMP/DIB, if the alpha channel is set to a
+         * non-zero value on any pixel we get a different format.
+         */
+        { D3DFMT_A8R8G8B8, NULL,
           { { D3D_OK, D3DFMT_A8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4252,9 +3780,9 @@
             { D3D_OK, D3DFMT_A8R8G8B8 },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-          }
-        },
-        { D3DFMT_X8R8G8B8, NULL, 0x00,
+          }, 0xff
+        },
+        { D3DFMT_X8R8G8B8, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8, },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4270,7 +3798,7 @@
          * D3DFMT_A8B8G8R8 doesn't get replaced with D3DFMT_X8B8G8R8
          * for BMP/DIB.
          */
-        { D3DFMT_A8B8G8R8, NULL, 0x00,
+        { D3DFMT_A8B8G8R8, NULL,
           { { D3D_OK, D3DFMT_A8B8G8R8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4281,7 +3809,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_X8B8G8R8, NULL, 0x00,
+        { D3DFMT_X8B8G8R8, NULL,
           { { D3D_OK, D3DFMT_X8B8G8R8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4292,7 +3820,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_R5G6B5, NULL, 0x00,
+        { D3DFMT_R5G6B5, NULL,
           { { D3D_OK, D3DFMT_R5G6B5, },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4303,7 +3831,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_X1R5G5B5, NULL, 0x00,
+        { D3DFMT_X1R5G5B5, NULL,
           { { D3D_OK, D3DFMT_X1R5G5B5, },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4314,7 +3842,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A1R5G5B5, NULL, 0x00,
+        { D3DFMT_A1R5G5B5, NULL,
           { { D3D_OK, D3DFMT_A1R5G5B5, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4325,7 +3853,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_R3G3B2, NULL, 0x00,
+        { D3DFMT_R3G3B2, NULL,
           { { D3D_OK, D3DFMT_P8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4336,7 +3864,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A8R3G3B2, NULL, 0x00,
+        { D3DFMT_A8R3G3B2, NULL,
           { { D3D_OK, D3DFMT_A8R3G3B2, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4347,7 +3875,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A4R4G4B4, NULL, 0x00,
+        { D3DFMT_A4R4G4B4, NULL,
           { { D3D_OK, D3DFMT_A4R4G4B4, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4358,7 +3886,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_X4R4G4B4, NULL, 0x00,
+        { D3DFMT_X4R4G4B4, NULL,
           { { D3D_OK, D3DFMT_X4R4G4B4, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4369,7 +3897,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A2R10G10B10, NULL, 0x00,
+        { D3DFMT_A2R10G10B10, NULL,
           { { D3D_OK, D3DFMT_A2R10G10B10, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4380,7 +3908,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A2B10G10R10, NULL, 0x00,
+        { D3DFMT_A2B10G10R10, NULL,
           { { D3D_OK, D3DFMT_A2B10G10R10, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4391,7 +3919,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A16B16G16R16, NULL, 0x00,
+        { D3DFMT_A16B16G16R16, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4402,7 +3930,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_G16R16, NULL, 0x00,
+        { D3DFMT_G16R16, NULL,
           { { D3D_OK, D3DFMT_G16R16, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4413,7 +3941,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A8, NULL, 0x00,
+        { D3DFMT_A8, NULL,
           { { D3D_OK, D3DFMT_A8R3G3B2, .todo_hr = TRUE },
             { D3DERR_INVALIDCALL },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4424,7 +3952,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A8L8, NULL, 0x00,
+        { D3DFMT_A8L8, NULL,
           { { D3D_OK, D3DFMT_A8L8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4435,7 +3963,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A4L4, NULL, 0x00,
+        { D3DFMT_A4L4, NULL,
           { { D3D_OK, D3DFMT_A4R4G4B4, .todo_format = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4452,7 +3980,7 @@
          * value acts as an index into this palette. Weird that there are L16
          * and A8L8 representations, but this is done in a unique way.
          */
-        { D3DFMT_L8, NULL, 0x00,
+        { D3DFMT_L8, NULL,
           { { D3D_OK, D3DFMT_P8, .todo_format = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4463,7 +3991,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_L16, NULL, 0x00,
+        { D3DFMT_L16, NULL,
           { { D3D_OK, D3DFMT_L16, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4474,7 +4002,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_DXT1, NULL, 0x00,
+        { D3DFMT_DXT1, NULL,
           { { D3D_OK, D3DFMT_A8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4485,7 +4013,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_DXT2, NULL, 0x00,
+        { D3DFMT_DXT2, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4496,7 +4024,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_DXT2, NULL, 0xff,
+        { D3DFMT_DXT2, NULL,
           { { D3D_OK, D3DFMT_A8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4505,9 +4033,9 @@
             { D3D_OK, D3DFMT_A8R8G8B8 },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-          },
-        },
-        { D3DFMT_DXT3, NULL, 0x00,
+          }, 0xff
+        },
+        { D3DFMT_DXT3, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4518,18 +4046,18 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_DXT3, NULL, 0xff,
-          { { D3D_OK, D3DFMT_A8R8G8B8 },
+        { D3DFMT_DXT4, NULL,
+          { { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
+            { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_DXT4, NULL, 0x00,
+        { D3DFMT_DXT5, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4540,40 +4068,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_DXT4, NULL, 0xff,
-          { { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_X8R8G8B8 },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-          },
-        },
-        { D3DFMT_DXT5, NULL, 0x00,
-          { { D3D_OK, D3DFMT_X8R8G8B8 },
-            { D3D_OK, D3DFMT_X8R8G8B8 },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_X8R8G8B8 },
-            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-          },
-        },
-        { D3DFMT_DXT5, NULL, 0xff,
-          { { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_X8R8G8B8 },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_A8R8G8B8 },
-            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-            { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-          },
-        },
-        { D3DFMT_R16F, NULL, 0x00,
+        { D3DFMT_R16F, NULL,
           { { D3D_OK, D3DFMT_G16R16, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4584,7 +4079,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_G16R16F, NULL, 0x00,
+        { D3DFMT_G16R16F, NULL,
           { { D3D_OK, D3DFMT_G16R16, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4595,7 +4090,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A16B16G16R16F, NULL, 0x00,
+        { D3DFMT_A16B16G16R16F, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4606,7 +4101,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_R32F, NULL, 0x00,
+        { D3DFMT_R32F, NULL,
           { { D3D_OK, D3DFMT_G16R16, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4617,7 +4112,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_G32R32F, NULL, 0x00,
+        { D3DFMT_G32R32F, NULL,
           { { D3D_OK, D3DFMT_G16R16, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_R8G8B8 },
@@ -4628,7 +4123,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A32B32G32R32F, NULL, 0x00,
+        { D3DFMT_A32B32G32R32F, NULL,
           { { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_X8R8G8B8 },
             { D3D_OK, D3DFMT_A8R8G8B8 },
@@ -4639,7 +4134,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_Q8W8V8U8, NULL, 0x00,
+        { D3DFMT_Q8W8V8U8, NULL,
           { { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
@@ -4650,7 +4145,7 @@
             { D3DERR_INVALIDCALL, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_V8U8, NULL, 0x00,
+        { D3DFMT_V8U8, NULL,
           { { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
@@ -4661,7 +4156,7 @@
             { D3DERR_INVALIDCALL, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_V16U16, NULL, 0x00,
+        { D3DFMT_V16U16, NULL,
           { { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
@@ -4672,7 +4167,7 @@
             { D3DERR_INVALIDCALL, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_X8L8V8U8, NULL, 0x00,
+        { D3DFMT_X8L8V8U8, NULL,
           { { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
@@ -4683,7 +4178,7 @@
             { D3DERR_INVALIDCALL, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_A2W10V10U10, NULL, 0x00,
+        { D3DFMT_A2W10V10U10, NULL,
           { { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
@@ -4694,7 +4189,7 @@
             { D3DERR_INVALIDCALL, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_Q16W16V16U16, NULL, 0x00,
+        { D3DFMT_Q16W16V16U16, NULL,
           { { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
             { D3DERR_INVALIDCALL },
@@ -4705,7 +4200,7 @@
             { D3DERR_INVALIDCALL, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_R8G8_B8G8, NULL, 0x00,
+        { D3DFMT_R8G8_B8G8, NULL,
           { { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
@@ -4716,7 +4211,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_G8R8_G8B8, NULL, 0x00,
+        { D3DFMT_G8R8_G8B8, NULL,
           { { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
@@ -4727,7 +4222,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_UYVY, NULL, 0x00,
+        { D3DFMT_UYVY, NULL,
           { { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
@@ -4738,7 +4233,7 @@
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
           },
         },
-        { D3DFMT_YUY2, NULL, 0x00,
+        { D3DFMT_YUY2, NULL,
           { { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_X8R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
@@ -4747,23 +4242,12 @@
             { D3D_OK, D3DFMT_R8G8B8, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
             { D3D_OK, D3DFMT_A32B32G32R32F, .todo_hr = TRUE },
-          },
-        },
-        { D3DFMT_CxV8U8, NULL, 0x00,
-          { { D3DERR_INVALIDCALL },
-            { D3DERR_INVALIDCALL },
-            { D3DERR_INVALIDCALL },
-            { D3DERR_INVALIDCALL },
-            { D3DERR_INVALIDCALL, .todo_hr = TRUE },
-            { D3DERR_INVALIDCALL },
-            { D3DERR_INVALIDCALL, .todo_hr = TRUE },
-            { D3DERR_INVALIDCALL, .todo_hr = TRUE },
           },
         },
     };
     IDirect3DSurface9 *surface;
     ID3DXBuffer *buffer;
-    unsigned int i, j;
+    uint32_t i, j;
     HRESULT hr;
 
     for (i = 0; i < ARRAY_SIZE(iff_tests); ++i)
@@ -4784,32 +4268,27 @@
             const uint32_t tmp_height = is_dxt_d3dformat(iff_tests[i].format) ? 1 : 4;
 
             for (j = 0; j < tmp_height; ++j)
-                memset(((uint8_t *)lockrect.pBits) + (j * lockrect.Pitch), iff_tests[i].init_pixel_value,
-                        lockrect.Pitch);
+                memset(((uint8_t *)lockrect.pBits) + (j * lockrect.Pitch), iff_tests[i].init_pixel_value, lockrect.Pitch);
             IDirect3DSurface9_UnlockRect(surface);
         }
 
         winetest_push_context("Test %u (%s)", i, debug_d3dformat(iff_tests[i].format));
-        for (j = 0; j < ARRAY_SIZE(test_iffs); ++j)
+        for (j = 0; j < ARRAY_SIZE(test_iff); ++j)
         {
-            winetest_push_context("File format %u (%s)", j, test_iffs[j].name);
+            winetest_push_context("File format %u (%s)", j, test_iff_str[j]);
 
             buffer = NULL;
-            hr = D3DXSaveSurfaceToFileInMemory(&buffer, test_iffs[j].format, surface, iff_tests[i].palette, NULL);
-            todo_wine_if(iff_tests[i].iff_expected[j].todo_hr) ok(hr == iff_tests[i].iff_expected[j].hr,
-                    "Unexpected hr %#lx.\n", hr);
+            hr = D3DXSaveSurfaceToFileInMemory(&buffer, test_iff[j], surface, iff_tests[i].palette, NULL);
+            todo_wine_if(iff_tests[i].iff_expected[j].todo_hr) ok(hr == iff_tests[i].iff_expected[j].hr, "Unexpected hr %#lx.\n", hr);
             if (SUCCEEDED(hr))
             {
                 D3DXIMAGE_INFO info = { 0 };
 
-                hr = D3DXGetImageInfoFromFileInMemory(ID3DXBuffer_GetBufferPointer(buffer),
-                        ID3DXBuffer_GetBufferSize(buffer), &info);
+                hr = D3DXGetImageInfoFromFileInMemory(ID3DXBuffer_GetBufferPointer(buffer), ID3DXBuffer_GetBufferSize(buffer), &info);
                 ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-                todo_wine_if(iff_tests[i].iff_expected[j].todo_format)
-                    ok(info.Format == iff_tests[i].iff_expected[j].format,
-                            "Unexpected image format %u (%s), expected %u (%s).\n", info.Format,
-                            debug_d3dformat(info.Format), iff_tests[i].iff_expected[j].format,
-                            debug_d3dformat(iff_tests[i].iff_expected[j].format));
+                todo_wine_if(iff_tests[i].iff_expected[j].todo_format) ok(info.Format == iff_tests[i].iff_expected[j].format,
+                        "Unexpected image format %u (%s), expected %u (%s).\n", info.Format, debug_d3dformat(info.Format),
+                        iff_tests[i].iff_expected[j].format, debug_d3dformat(iff_tests[i].iff_expected[j].format));
                 ID3DXBuffer_Release(buffer);
             }
             winetest_pop_context();
@@ -5134,160 +4613,6 @@
     IDirect3DSurface9_Release(surface);
 }
 
-static void test_image_filters(void)
-{
-    static const struct
-    {
-        DWORD filter;
-        const char *name;
-    }
-    test_filters[] =
-    {
-        { D3DX_FILTER_POINT, "D3DX_FILTER_POINT" },
-        { D3DX_FILTER_LINEAR, "D3DX_FILTER_LINEAR" },
-        { D3DX_FILTER_TRIANGLE, "D3DX_FILTER_TRIANGLE" },
-        { D3DX_FILTER_BOX, "D3DX_FILTER_BOX" },
-    };
-    static const struct
-    {
-        const RECT src_rect;
-        const void *src_data;
-        D3DFORMAT format;
-
-        unsigned int dst_width;
-        unsigned int dst_height;
-        struct
-        {
-            const void *expected_dst_data;
-            const void *expected_dst_data_32;
-            BOOL todo;
-        }
-        filter_expected[4];
-    } tests[] =
-    {
-        {
-            { 0, 0, 16, 16 }, a8r8g8b8_16_16, D3DFMT_A8R8G8B8, 8, 8,
-            {
-                { a8r8g8b8_16_16_point_filter_8_8 },
-                { a8r8g8b8_16_16_linear_filter_8_8, .todo = TRUE },
-                { a8r8g8b8_16_16_triangle_filter_8_8, a8r8g8b8_16_16_triangle_filter_8_8_32bit, .todo = TRUE },
-                /* Linear and box filters match. */
-                { a8r8g8b8_16_16_linear_filter_8_8 },
-            },
-        },
-        {
-            { 0, 0, 16, 16 }, a8r8g8b8_16_16, D3DFMT_A8R8G8B8, 7, 7,
-            {
-                { a8r8g8b8_16_16_point_filter_7_7, 0, },
-                { a8r8g8b8_16_16_linear_filter_7_7, .todo = TRUE },
-                { a8r8g8b8_16_16_triangle_filter_7_7, .todo = TRUE },
-                /*
-                 * If scaling down to a size that isn't a power of two, the
-                 * box filter is replaced with the triangle filter.
-                 */
-                { a8r8g8b8_16_16_triangle_filter_7_7, .todo = TRUE },
-            },
-        },
-        {
-            { 0, 0, 6, 6 }, a8r8g8b8_6_6, D3DFMT_A8R8G8B8, 3, 3,
-            {
-                { a8r8g8b8_6_6_point_filter_3_3, 0, },
-                { a8r8g8b8_6_6_linear_filter_3_3, .todo = TRUE },
-                { a8r8g8b8_6_6_triangle_filter_3_3, .todo = TRUE },
-                /*
-                 * If scaling down to a size that is half but isn't a power of
-                 * two, the box filter is replaced with the linear filter.
-                 */
-                { a8r8g8b8_6_6_linear_filter_3_3, .todo = TRUE },
-            },
-        },
-        {
-            { 0, 0, 6, 6 }, a8r8g8b8_6_6, D3DFMT_A8R8G8B8, 4, 4,
-            {
-                { a8r8g8b8_6_6_point_filter_4_4, 0, },
-                { a8r8g8b8_6_6_linear_filter_4_4, .todo = TRUE },
-                { a8r8g8b8_6_6_triangle_filter_4_4, .todo = TRUE },
-                /*
-                 * If scaling down to a size that is not half and isn't a power of
-                 * two, the box filter is replaced with the triangle filter.
-                 */
-                { a8r8g8b8_6_6_triangle_filter_4_4, .todo = TRUE },
-            },
-        },
-        {
-            { 0, 0, 4, 4 }, a8r8g8b8_4_4, D3DFMT_A8R8G8B8, 12, 12,
-            {
-                { a8r8g8b8_4_4_point_filter_12_12, .todo = TRUE },
-                { a8r8g8b8_4_4_linear_filter_12_12, .todo = TRUE },
-                { a8r8g8b8_4_4_triangle_filter_12_12, .todo = TRUE },
-                { a8r8g8b8_4_4_triangle_filter_12_12, .todo = TRUE },
-            },
-        },
-        {
-            { 0, 0, 4, 4 }, a8r8g8b8_4_4, D3DFMT_A8R8G8B8, 8, 8,
-            {
-                { a8r8g8b8_4_4_point_filter_8_8 },
-                /*
-                 * When scaling up by 2, linear and triangle filters behave
-                 * the same.
-                 */
-                { a8r8g8b8_4_4_linear_filter_8_8, .todo = TRUE },
-                { a8r8g8b8_4_4_linear_filter_8_8, .todo = TRUE },
-                { a8r8g8b8_4_4_linear_filter_8_8, .todo = TRUE },
-            },
-        },
-    };
-    const uint8_t *expected_dst;
-    IDirect3DDevice9 *device;
-    D3DLOCKED_RECT lock_rect;
-    IDirect3DSurface9 *surf;
-    uint32_t src_pitch;
-    unsigned int i, j;
-    HRESULT hr;
-    HWND hwnd;
-
-    if (!(device = create_device(&hwnd)))
-        return;
-
-    for (i = 0; i < ARRAY_SIZE(tests); ++i)
-    {
-        const unsigned int fmt_bpp = get_bpp_for_d3dformat(tests[i].format);
-
-        winetest_push_context("Test %u", i);
-
-        hr = IDirect3DDevice9_CreateOffscreenPlainSurface(device, tests[i].dst_width, tests[i].dst_height,
-                tests[i].format, D3DPOOL_SCRATCH, &surf, NULL);
-        ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-
-        src_pitch = fmt_bpp * (tests[i].src_rect.right - tests[i].src_rect.left);
-        for (j = 0; j < ARRAY_SIZE(test_filters); ++j)
-        {
-            winetest_push_context("Filter %s", test_filters[j].name);
-
-            hr = D3DXLoadSurfaceFromMemory(surf, NULL, NULL, tests[i].src_data, tests[i].format,
-                    src_pitch, NULL, &tests[i].src_rect, test_filters[j].filter, 0);
-            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
-            if (sizeof(void *) == 4 && tests[i].filter_expected[j].expected_dst_data_32)
-                expected_dst = tests[i].filter_expected[j].expected_dst_data_32;
-            else
-                expected_dst = tests[i].filter_expected[j].expected_dst_data;
-
-            IDirect3DSurface9_LockRect(surf, &lock_rect, NULL, D3DLOCK_READONLY);
-            todo_wine_if(tests[i].filter_expected[j].todo) check_test_readback(lock_rect.pBits, lock_rect.Pitch, 0,
-                    expected_dst, tests[i].dst_width, tests[i].dst_height, 1, tests[i].format, 0);
-            IDirect3DSurface9_UnlockRect(surf);
-
-            winetest_pop_context();
-        }
-
-        check_release((IUnknown *)surf, 0);
-        winetest_pop_context();
-    }
-
-    check_release((IUnknown *)device, 0);
-    DestroyWindow(hwnd);
-}
-
 START_TEST(surface)
 {
     HWND wnd;
@@ -5328,7 +4653,4 @@
     check_release((IUnknown*)device, 0);
     check_release((IUnknown*)d3d, 0);
     DestroyWindow(wnd);
-
-    test_color_key();
-    test_image_filters();
 }
