diff --git a/dlls/mshtml/xmlhttprequest.c b/dlls/mshtml/xmlhttprequest.c
--- a/dlls/mshtml/xmlhttprequest.c
+++ b/dlls/mshtml/xmlhttprequest.c
@@ -29,14 +29,26 @@
 
 #include "mshtml_private.h"
 #include "htmlevent.h"
-#include "binding.h"
 #include "mshtmdid.h"
 #include "initguid.h"
 #include "msxml6.h"
 #include "objsafe.h"
-#include "shlwapi.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(mshtml);
+
+typedef struct {
+    DispatchEx dispex;
+    IHTMLXMLHttpRequestFactory IHTMLXMLHttpRequestFactory_iface;
+
+    HTMLInnerWindow *window;
+} HTMLXMLHttpRequestFactory;
+
+typedef struct {
+    DispatchEx dispex;
+    IHTMLXDomainRequestFactory IHTMLXDomainRequestFactory_iface;
+
+    HTMLInnerWindow *window;
+} HTMLXDomainRequestFactory;
 
 static HRESULT bstr_to_nsacstr(BSTR bstr, nsACString *str)
 {
@@ -129,13 +141,19 @@
 typedef struct {
     nsIDOMEventListener nsIDOMEventListener_iface;
     LONG ref;
-    struct xhr *xhr;
+    HTMLXMLHttpRequest *xhr;
 } XMLHttpReqEventListener;
 
-struct xhr {
+struct HTMLXMLHttpRequest {
     EventTarget event_target;
+    IHTMLXMLHttpRequest IHTMLXMLHttpRequest_iface;
+    IHTMLXMLHttpRequest2 IHTMLXMLHttpRequest2_iface;
+    IWineXMLHttpRequestPrivate IWineXMLHttpRequestPrivate_iface;
+    IProvideClassInfo2 IProvideClassInfo2_iface;
+    IHTMLXDomainRequest IHTMLXDomainRequest_iface;  /* only for XDomainRequests */
     LONG task_magic;
     LONG ready_state;
+    document_type_t doctype_override;
     response_type_t response_type;
     BOOLEAN synchronous;
     DWORD magic;
@@ -147,19 +165,6 @@
     DOMEvent *pending_progress_event;
 };
 
-struct HTMLXMLHttpRequest {
-    struct xhr xhr;
-    IHTMLXMLHttpRequest IHTMLXMLHttpRequest_iface;
-    IHTMLXMLHttpRequest2 IHTMLXMLHttpRequest2_iface;
-    IWineXMLHttpRequestPrivate IWineXMLHttpRequestPrivate_iface;
-    IProvideClassInfo2 IProvideClassInfo2_iface;
-};
-
-typedef struct {
-    struct xhr xhr;
-    IHTMLXDomainRequest IHTMLXDomainRequest_iface;
-} HTMLXDomainRequest;
-
 static void detach_xhr_event_listener(XMLHttpReqEventListener *event_listener)
 {
     nsIDOMEventTarget *event_target;
@@ -183,7 +188,7 @@
     nsIDOMEventListener_Release(&event_listener->nsIDOMEventListener_iface);
 }
 
-static void synthesize_pending_events(struct xhr *xhr)
+static void synthesize_pending_events(HTMLXMLHttpRequest *xhr)
 {
     DWORD magic = xhr->pending_events_magic;
     UINT16 ready_state = xhr->ready_state;
@@ -211,7 +216,7 @@
         xhr->ready_state = max(xhr->ready_state, READYSTATE_INTERACTIVE);
     }
 
-    IWineJSDispatchHost_AddRef(&xhr->event_target.dispex.IWineJSDispatchHost_iface);
+    IHTMLXMLHttpRequest_AddRef(&xhr->IHTMLXMLHttpRequest_iface);
 
     send_loadend = send_load = (xhr->ready_state != ready_state && ready_state == READYSTATE_COMPLETE);
     for(;;) {
@@ -266,14 +271,14 @@
     }
 
 ret:
-    IWineJSDispatchHost_Release(&xhr->event_target.dispex.IWineJSDispatchHost_iface);
-}
-
-static nsresult sync_xhr_send(struct xhr *xhr, nsIVariant *nsbody)
+    IHTMLXMLHttpRequest_Release(&xhr->IHTMLXMLHttpRequest_iface);
+}
+
+static nsresult sync_xhr_send(HTMLXMLHttpRequest *xhr, nsIVariant *nsbody)
 {
     thread_data_t *thread_data = get_thread_data(TRUE);
+    HTMLXMLHttpRequest *prev_blocking_xhr;
     HTMLInnerWindow *window = xhr->window;
-    struct xhr *prev_blocking_xhr;
     unsigned prev_tasks_locked;
     nsresult nsres;
 
@@ -333,7 +338,7 @@
 
 struct pending_xhr_events_task {
     event_task_t header;
-    struct xhr *xhr;
+    HTMLXMLHttpRequest *xhr;
 };
 
 static void pending_xhr_events_proc(event_task_t *_task)
@@ -401,7 +406,7 @@
 static nsresult NSAPI XMLHttpReqEventListener_HandleEvent(nsIDOMEventListener *iface, nsIDOMEvent *nsevent)
 {
     XMLHttpReqEventListener *This = impl_from_nsIDOMEventListener(iface);
-    struct xhr *blocking_xhr = NULL;
+    HTMLXMLHttpRequest *blocking_xhr = NULL;
     thread_data_t *thread_data;
     compat_mode_t compat_mode;
     LONG ready_state;
@@ -529,7 +534,7 @@
 }
 
 DISPEX_IDISPATCH_IMPL(HTMLXMLHttpRequest, IHTMLXMLHttpRequest,
-                      impl_from_IHTMLXMLHttpRequest(iface)->xhr.event_target.dispex)
+                      impl_from_IHTMLXMLHttpRequest(iface)->event_target.dispex)
 
 static HRESULT WINAPI HTMLXMLHttpRequest_get_readyState(IHTMLXMLHttpRequest *iface, LONG *p)
 {
@@ -539,7 +544,7 @@
 
     if(!p)
         return E_POINTER;
-    *p = This->xhr.ready_state;
+    *p = This->ready_state;
     return S_OK;
 }
 
@@ -550,31 +555,25 @@
     return E_NOTIMPL;
 }
 
-static HRESULT WINAPI get_response_text(struct xhr *xhr, BSTR *p)
-{
+static HRESULT WINAPI HTMLXMLHttpRequest_get_responseText(IHTMLXMLHttpRequest *iface, BSTR *p)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXMLHttpRequest(iface);
     nsAString nsstr;
     nsresult nsres;
 
+    TRACE("(%p)->(%p)\n", This, p);
+
     if(!p)
         return E_POINTER;
 
-    if(xhr->ready_state < READYSTATE_INTERACTIVE) {
+    if(This->ready_state < READYSTATE_INTERACTIVE) {
         *p = NULL;
         return S_OK;
     }
 
     nsAString_Init(&nsstr, NULL);
-    nsres = nsIXMLHttpRequest_GetResponseText(xhr->nsxhr, &nsstr);
+    nsres = nsIXMLHttpRequest_GetResponseText(This->nsxhr, &nsstr);
     return return_nsstr(nsres, &nsstr, p);
-}
-
-static HRESULT WINAPI HTMLXMLHttpRequest_get_responseText(IHTMLXMLHttpRequest *iface, BSTR *p)
-{
-    HTMLXMLHttpRequest *This = impl_from_IHTMLXMLHttpRequest(iface);
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    return get_response_text(&This->xhr, p);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_get_responseXML(IHTMLXMLHttpRequest *iface, IDispatch **p)
@@ -588,17 +587,20 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    if(This->xhr.ready_state < READYSTATE_COMPLETE) {
+    if(This->ready_state < READYSTATE_COMPLETE) {
         *p = NULL;
         return S_OK;
     }
 
-    if(dispex_compat_mode(&This->xhr.event_target.dispex) >= COMPAT_MODE_IE10) {
+    if(dispex_compat_mode(&This->event_target.dispex) >= COMPAT_MODE_IE10) {
+        nsACString header, nscstr;
+        document_type_t doctype;
         HTMLDocumentNode *doc;
         nsIDOMDocument *nsdoc;
+        const char *type;
         nsresult nsres;
 
-        nsres = nsIXMLHttpRequest_GetResponseXML(This->xhr.nsxhr, &nsdoc);
+        nsres = nsIXMLHttpRequest_GetResponseXML(This->nsxhr, &nsdoc);
         if(NS_FAILED(nsres))
             return map_nsresult(nsres);
         if(!nsdoc) {
@@ -606,15 +608,35 @@
             return S_OK;
         }
 
-        if(!This->xhr.window->base.outer_window || !This->xhr.window->base.outer_window->browser)
+        if(!This->window->base.outer_window || !This->window->base.outer_window->browser)
             hres = E_UNEXPECTED;
-        else
-            hres = create_document_node(nsdoc, This->xhr.window->base.outer_window->browser, NULL, This->xhr.window,
-                                        dispex_compat_mode(&This->xhr.window->event_target.dispex), &doc);
+        else {
+            if(This->doctype_override != DOCTYPE_INVALID)
+                doctype = This->doctype_override;
+            else {
+                doctype = DOCTYPE_XML;
+                nsACString_InitDepend(&header, "Content-Type");
+                nsACString_InitDepend(&nscstr, NULL);
+                nsres = nsIXMLHttpRequest_GetResponseHeader(This->nsxhr, &header, &nscstr);
+                nsACString_Finish(&header);
+                if(NS_SUCCEEDED(nsres)) {
+                    nsACString_GetData(&nscstr, &type);
+                    if(!stricmp(type, "application/xhtml+xml"))
+                        doctype = DOCTYPE_XHTML;
+                    else if(!stricmp(type, "image/svg+xml"))
+                        doctype = DOCTYPE_SVG;
+                }
+                nsACString_Finish(&nscstr);
+            }
+            hres = create_document_node(nsdoc, This->window->base.outer_window->browser, NULL, This->window,
+                                        doctype, dispex_compat_mode(&This->window->event_target.dispex), &doc);
+        }
         nsIDOMDocument_Release(nsdoc);
         if(FAILED(hres))
             return hres;
 
+        /* make sure dispex info is initialized */
+        dispex_compat_mode(&doc->node.event_target.dispex);
         *p = (IDispatch*)&doc->IHTMLDocument2_iface;
         return S_OK;
     }
@@ -659,12 +681,12 @@
     if(!p)
         return E_POINTER;
 
-    if(This->xhr.ready_state < READYSTATE_LOADED) {
+    if(This->ready_state < READYSTATE_LOADED) {
         *p = 0;
         return E_FAIL;
     }
 
-    nsres = nsIXMLHttpRequest_GetStatus(This->xhr.nsxhr, &val);
+    nsres = nsIXMLHttpRequest_GetStatus(This->nsxhr, &val);
     if(NS_FAILED(nsres)) {
         ERR("nsIXMLHttpRequest_GetStatus failed: %08lx\n", nsres);
         return E_FAIL;
@@ -687,13 +709,13 @@
     if(!p)
         return E_POINTER;
 
-    if(This->xhr.ready_state < READYSTATE_LOADED) {
+    if(This->ready_state < READYSTATE_LOADED) {
         *p = NULL;
         return E_FAIL;
     }
 
     nsACString_Init(&nscstr, NULL);
-    nsres = nsIXMLHttpRequest_GetStatusText(This->xhr.nsxhr, &nscstr);
+    nsres = nsIXMLHttpRequest_GetStatusText(This->nsxhr, &nscstr);
     return return_nscstr(nsres, &nscstr, p);
 }
 
@@ -703,7 +725,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
 
-    return set_event_handler(&This->xhr.event_target, EVENTID_READYSTATECHANGE, &v);
+    return set_event_handler(&This->event_target, EVENTID_READYSTATECHANGE, &v);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_get_onreadystatechange(IHTMLXMLHttpRequest *iface, VARIANT *p)
@@ -712,74 +734,30 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return get_event_handler(&This->xhr.event_target, EVENTID_READYSTATECHANGE, p);
-}
-
-static HRESULT WINAPI xhr_abort(struct xhr *xhr)
-{
-    DWORD prev_magic = xhr->magic;
+    return get_event_handler(&This->event_target, EVENTID_READYSTATECHANGE, p);
+}
+
+static HRESULT WINAPI HTMLXMLHttpRequest_abort(IHTMLXMLHttpRequest *iface)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXMLHttpRequest(iface);
+    DWORD prev_magic = This->magic;
     UINT16 ready_state;
     nsresult nsres;
 
-    xhr->magic++;
-    nsres = nsIXMLHttpRequest_SlowAbort(xhr->nsxhr);
+    TRACE("(%p)->()\n", This);
+
+    This->magic++;
+    nsres = nsIXMLHttpRequest_SlowAbort(This->nsxhr);
     if(NS_FAILED(nsres)) {
         ERR("nsIXMLHttpRequest_SlowAbort failed: %08lx\n", nsres);
-        xhr->magic = prev_magic;
+        This->magic = prev_magic;
         return E_FAIL;
     }
 
     /* Gecko changed to READYSTATE_UNINITIALIZED if it did abort */
-    nsres = nsIXMLHttpRequest_GetReadyState(xhr->nsxhr, &ready_state);
+    nsres = nsIXMLHttpRequest_GetReadyState(This->nsxhr, &ready_state);
     if(NS_SUCCEEDED(nsres))
-        xhr->ready_state = ready_state;
-    return S_OK;
-}
-
-static HRESULT WINAPI HTMLXMLHttpRequest_abort(IHTMLXMLHttpRequest *iface)
-{
-    HTMLXMLHttpRequest *This = impl_from_IHTMLXMLHttpRequest(iface);
-
-    TRACE("(%p)->()\n", This);
-
-    return xhr_abort(&This->xhr);
-}
-
-static HRESULT WINAPI xhr_open(struct xhr *xhr, BSTR bstrMethod, BSTR bstrUrl, BOOL async, nsAString *user, nsAString *password, unsigned opt_argc)
-{
-    BOOLEAN prev_synchronous;
-    nsACString method, url;
-    DWORD prev_magic;
-    nsresult nsres;
-    HRESULT hres;
-
-    hres = bstr_to_nsacstr(bstrMethod, &method);
-    if(FAILED(hres))
-        return hres;
-    hres = bstr_to_nsacstr(bstrUrl, &url);
-    if(FAILED(hres)) {
-        nsACString_Finish(&method);
-        return hres;
-    }
-
-    /* Set this here, Gecko dispatches nested sync XHR readyState changes for OPENED (see HandleEvent) */
-    prev_magic = xhr->magic;
-    prev_synchronous = xhr->synchronous;
-    xhr->synchronous = !async;
-    xhr->magic++;
-
-    nsres = nsIXMLHttpRequest_Open(xhr->nsxhr, &method, &url, async, user, password, opt_argc);
-
-    nsACString_Finish(&method);
-    nsACString_Finish(&url);
-
-    if(NS_FAILED(nsres)) {
-        ERR("nsIXMLHttpRequest_Open failed: %08lx\n", nsres);
-        xhr->magic = prev_magic;
-        xhr->synchronous = prev_synchronous;
-        return E_FAIL;
-    }
-
+        This->ready_state = ready_state;
     return S_OK;
 }
 
@@ -808,8 +786,12 @@
 static HRESULT WINAPI HTMLXMLHttpRequest_open(IHTMLXMLHttpRequest *iface, BSTR bstrMethod, BSTR bstrUrl, VARIANT varAsync, VARIANT varUser, VARIANT varPassword)
 {
     HTMLXMLHttpRequest *This = impl_from_IHTMLXMLHttpRequest(iface);
+    BOOLEAN prev_synchronous;
     nsAString user, password;
+    nsACString method, url;
     unsigned opt_argc = 1;
+    DWORD prev_magic;
+    nsresult nsres;
     HRESULT hres;
 
     TRACE("(%p)->(%s %s %s %s %s)\n", This, debugstr_w(bstrMethod), debugstr_w(bstrUrl), debugstr_variant(&varAsync), debugstr_variant(&varUser), debugstr_variant(&varPassword));
@@ -832,23 +814,56 @@
         return hres;
     }
 
+    hres = bstr_to_nsacstr(bstrMethod, &method);
+    if(FAILED(hres)) {
+        nsAString_Finish(&user);
+        nsAString_Finish(&password);
+        return hres;
+    }
+    hres = bstr_to_nsacstr(bstrUrl, &url);
+    if(FAILED(hres)) {
+        nsAString_Finish(&user);
+        nsAString_Finish(&password);
+        nsACString_Finish(&method);
+        return hres;
+    }
+
+    /* Set this here, Gecko dispatches nested sync XHR readyState changes for OPENED (see HandleEvent) */
+    prev_magic = This->magic;
+    prev_synchronous = This->synchronous;
+    This->synchronous = !V_BOOL(&varAsync);
+    This->magic++;
+
     if(V_VT(&varPassword) != VT_EMPTY && V_VT(&varPassword) != VT_ERROR)
         opt_argc += 2;
     else if(V_VT(&varUser) != VT_EMPTY && V_VT(&varUser) != VT_ERROR)
         opt_argc += 1;
-
-    hres = xhr_open(&This->xhr, bstrMethod, bstrUrl, !!V_BOOL(&varAsync), &user, &password, opt_argc);
+    nsres = nsIXMLHttpRequest_Open(This->nsxhr, &method, &url, !!V_BOOL(&varAsync), &user, &password, opt_argc);
+
+    nsACString_Finish(&method);
+    nsACString_Finish(&url);
     nsAString_Finish(&user);
     nsAString_Finish(&password);
-    return hres;
-}
-
-static HRESULT WINAPI xhr_send(struct xhr *xhr, VARIANT *varBody)
-{
+
+    if(NS_FAILED(nsres)) {
+        ERR("nsIXMLHttpRequest_Open failed: %08lx\n", nsres);
+        This->magic = prev_magic;
+        This->synchronous = prev_synchronous;
+        return E_FAIL;
+    }
+
+    return S_OK;
+}
+
+static HRESULT WINAPI HTMLXMLHttpRequest_send(IHTMLXMLHttpRequest *iface, VARIANT varBody)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXMLHttpRequest(iface);
     nsIWritableVariant *nsbody = NULL;
     nsresult nsres = NS_OK;
 
-    switch(V_VT(varBody)) {
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&varBody));
+
+    switch(V_VT(&varBody)) {
     case VT_NULL:
     case VT_EMPTY:
     case VT_ERROR:
@@ -860,21 +875,21 @@
         if(!nsbody)
             return E_OUTOFMEMORY;
 
-        nsAString_InitDepend(&nsstr, V_BSTR(varBody));
+        nsAString_InitDepend(&nsstr, V_BSTR(&varBody));
         nsres = nsIWritableVariant_SetAsAString(nsbody, &nsstr);
         nsAString_Finish(&nsstr);
         break;
     }
     default:
-        FIXME("unsupported body type %s\n", debugstr_variant(varBody));
+        FIXME("unsupported body type %s\n", debugstr_variant(&varBody));
         return E_NOTIMPL;
     }
 
     if(NS_SUCCEEDED(nsres)) {
-        if(xhr->synchronous)
-            nsres = sync_xhr_send(xhr, (nsIVariant*)nsbody);
+        if(This->synchronous)
+            nsres = sync_xhr_send(This, (nsIVariant*)nsbody);
         else
-            nsres = nsIXMLHttpRequest_Send(xhr->nsxhr, (nsIVariant*)nsbody);
+            nsres = nsIXMLHttpRequest_Send(This->nsxhr, (nsIVariant*)nsbody);
     }
 
     if(nsbody)
@@ -887,15 +902,6 @@
     return S_OK;
 }
 
-static HRESULT WINAPI HTMLXMLHttpRequest_send(IHTMLXMLHttpRequest *iface, VARIANT varBody)
-{
-    HTMLXMLHttpRequest *This = impl_from_IHTMLXMLHttpRequest(iface);
-
-    TRACE("(%p)->(%s)\n", This, debugstr_variant(&varBody));
-
-    return xhr_send(&This->xhr, &varBody);
-}
-
 static HRESULT WINAPI HTMLXMLHttpRequest_getAllResponseHeaders(IHTMLXMLHttpRequest *iface, BSTR *p)
 {
     HTMLXMLHttpRequest *This = impl_from_IHTMLXMLHttpRequest(iface);
@@ -907,13 +913,13 @@
     if(!p)
         return E_POINTER;
 
-    if(This->xhr.ready_state < READYSTATE_LOADED) {
+    if(This->ready_state < READYSTATE_LOADED) {
         *p = NULL;
         return E_FAIL;
     }
 
     nsACString_Init(&nscstr, NULL);
-    nsres = nsIXMLHttpRequest_GetAllResponseHeaders(This->xhr.nsxhr, &nscstr);
+    nsres = nsIXMLHttpRequest_GetAllResponseHeaders(This->nsxhr, &nscstr);
     return return_nscstr(nsres, &nscstr, p);
 }
 
@@ -930,7 +936,7 @@
     if(!bstrHeader)
         return E_INVALIDARG;
 
-    if(This->xhr.ready_state < READYSTATE_LOADED) {
+    if(This->ready_state < READYSTATE_LOADED) {
         *p = NULL;
         return E_FAIL;
     }
@@ -939,7 +945,7 @@
     nsACString_InitDepend(&header, cstr);
     nsACString_Init(&ret, NULL);
 
-    nsres = nsIXMLHttpRequest_GetResponseHeader(This->xhr.nsxhr, &header, &ret);
+    nsres = nsIXMLHttpRequest_GetResponseHeader(This->nsxhr, &header, &ret);
 
     nsACString_Finish(&header);
     free(cstr);
@@ -967,7 +973,7 @@
 
     nsACString_InitDepend(&header, header_u);
     nsACString_InitDepend(&value, value_u);
-    nsres = nsIXMLHttpRequest_SetRequestHeader(This->xhr.nsxhr, &header, &value);
+    nsres = nsIXMLHttpRequest_SetRequestHeader(This->nsxhr, &header, &value);
     nsACString_Finish(&header);
     nsACString_Finish(&value);
     free(header_u);
@@ -1010,7 +1016,7 @@
 }
 
 DISPEX_IDISPATCH_IMPL(HTMLXMLHttpRequest2, IHTMLXMLHttpRequest2,
-                      impl_from_IHTMLXMLHttpRequest2(iface)->xhr.event_target.dispex)
+                      impl_from_IHTMLXMLHttpRequest2(iface)->event_target.dispex)
 
 static HRESULT WINAPI HTMLXMLHttpRequest2_put_timeout(IHTMLXMLHttpRequest2 *iface, LONG v)
 {
@@ -1020,7 +1026,7 @@
 
     if(v < 0)
         return E_INVALIDARG;
-    return map_nsresult(nsIXMLHttpRequest_SetTimeout(This->xhr.nsxhr, v));
+    return map_nsresult(nsIXMLHttpRequest_SetTimeout(This->nsxhr, v));
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest2_get_timeout(IHTMLXMLHttpRequest2 *iface, LONG *p)
@@ -1034,7 +1040,7 @@
     if(!p)
         return E_POINTER;
 
-    nsres = nsIXMLHttpRequest_GetTimeout(This->xhr.nsxhr, &timeout);
+    nsres = nsIXMLHttpRequest_GetTimeout(This->nsxhr, &timeout);
     *p = timeout;
     return map_nsresult(nsres);
 }
@@ -1045,7 +1051,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
 
-    return set_event_handler(&This->xhr.event_target, EVENTID_TIMEOUT, &v);
+    return set_event_handler(&This->event_target, EVENTID_TIMEOUT, &v);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest2_get_ontimeout(IHTMLXMLHttpRequest2 *iface, VARIANT *p)
@@ -1054,7 +1060,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return get_event_handler(&This->xhr.event_target, EVENTID_TIMEOUT, p);
+    return get_event_handler(&This->event_target, EVENTID_TIMEOUT, p);
 }
 
 static const IHTMLXMLHttpRequest2Vtbl HTMLXMLHttpRequest2Vtbl = {
@@ -1077,7 +1083,7 @@
 }
 
 DISPEX_IDISPATCH_IMPL(HTMLXMLHttpRequest_private, IWineXMLHttpRequestPrivate,
-                      impl_from_IWineXMLHttpRequestPrivate(iface)->xhr.event_target.dispex)
+                      impl_from_IWineXMLHttpRequestPrivate(iface)->event_target.dispex)
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_get_response(IWineXMLHttpRequestPrivate *iface, VARIANT *p)
 {
@@ -1089,14 +1095,14 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    if(This->xhr.response_obj) {
+    if(This->response_obj) {
         V_VT(p) = VT_DISPATCH;
-        V_DISPATCH(p) = This->xhr.response_obj;
-        IDispatch_AddRef(This->xhr.response_obj);
+        V_DISPATCH(p) = This->response_obj;
+        IDispatch_AddRef(This->response_obj);
         return S_OK;
     }
 
-    switch(This->xhr.response_type) {
+    switch(This->response_type) {
     case response_type_empty:
     case response_type_text:
         hres = IHTMLXMLHttpRequest_get_responseText(&This->IHTMLXMLHttpRequest_iface, &V_BSTR(p));
@@ -1110,17 +1116,17 @@
 
     case response_type_arraybuf:
     case response_type_blob:
-        if(This->xhr.ready_state < READYSTATE_COMPLETE) {
+        if(This->ready_state < READYSTATE_COMPLETE) {
             V_VT(p) = VT_EMPTY;
             break;
         }
-        nsres = nsIXMLHttpRequest_GetResponseBuffer(This->xhr.nsxhr, NULL, 0, &buf_size);
+        nsres = nsIXMLHttpRequest_GetResponseBuffer(This->nsxhr, NULL, 0, &buf_size);
         assert(nsres == NS_OK);
 
-        if(This->xhr.response_type == response_type_arraybuf) {
-            hres = IWineJScript_CreateArrayBuffer(This->xhr.window->jscript, buf_size, (IWineJSDispatch**)&This->xhr.response_obj, &buf);
+        if(This->response_type == response_type_arraybuf) {
+            hres = IWineJScript_CreateArrayBuffer(This->window->jscript, buf_size, (IWineJSDispatch**)&This->response_obj, &buf);
             if(SUCCEEDED(hres)) {
-                nsres = nsIXMLHttpRequest_GetResponseBuffer(This->xhr.nsxhr, buf, buf_size, &buf_size);
+                nsres = nsIXMLHttpRequest_GetResponseBuffer(This->nsxhr, buf, buf_size, &buf_size);
                 assert(nsres == NS_OK);
             }
             break;
@@ -1137,10 +1143,10 @@
         assert(0);
     }
 
-    if(SUCCEEDED(hres) && This->xhr.response_obj) {
+    if(SUCCEEDED(hres) && This->response_obj) {
         V_VT(p) = VT_DISPATCH;
-        V_DISPATCH(p) = This->xhr.response_obj;
-        IDispatch_AddRef(This->xhr.response_obj);
+        V_DISPATCH(p) = This->response_obj;
+        IDispatch_AddRef(This->response_obj);
     }
     return hres;
 }
@@ -1154,7 +1160,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_w(v));
 
-    if(This->xhr.ready_state < READYSTATE_LOADING || This->xhr.ready_state > READYSTATE_INTERACTIVE) {
+    if(This->ready_state < READYSTATE_LOADING || This->ready_state > READYSTATE_INTERACTIVE) {
         /* FIXME: Return InvalidStateError */
         return E_FAIL;
     }
@@ -1166,12 +1172,12 @@
         return S_OK;
 
     nsAString_InitDepend(&nsstr, response_type_desc[i].nsxhr_str);
-    nsres = nsIXMLHttpRequest_SetResponseType(This->xhr.nsxhr, &nsstr);
+    nsres = nsIXMLHttpRequest_SetResponseType(This->nsxhr, &nsstr);
     nsAString_Finish(&nsstr);
     if(NS_FAILED(nsres))
         return map_nsresult(nsres);
 
-    This->xhr.response_type = i;
+    This->response_type = i;
     return S_OK;
 }
 
@@ -1181,7 +1187,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    *p = SysAllocString(response_type_desc[This->xhr.response_type].str);
+    *p = SysAllocString(response_type_desc[This->response_type].str);
     return *p ? S_OK : E_OUTOFMEMORY;
 }
 
@@ -1200,7 +1206,7 @@
 
     TRACE("(%p)->(%x)\n", This, v);
 
-    return map_nsresult(nsIXMLHttpRequest_SetWithCredentials(This->xhr.nsxhr, !!v));
+    return map_nsresult(nsIXMLHttpRequest_SetWithCredentials(This->nsxhr, !!v));
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_get_withCredentials(IWineXMLHttpRequestPrivate *iface, VARIANT_BOOL *p)
@@ -1211,7 +1217,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    nsres = nsIXMLHttpRequest_GetWithCredentials(This->xhr.nsxhr, &b);
+    nsres = nsIXMLHttpRequest_GetWithCredentials(This->nsxhr, &b);
     if(NS_FAILED(nsres))
         return map_nsresult(nsres);
     *p = b ? VARIANT_TRUE : VARIANT_FALSE;
@@ -1222,6 +1228,7 @@
 {
     HTMLXMLHttpRequest *This = impl_from_IWineXMLHttpRequestPrivate(iface);
     static const WCHAR generic_type[] = L"application/octet-stream";
+    document_type_t doctype = DOCTYPE_XML;
     const WCHAR *type = NULL;
     WCHAR *lowercase = NULL;
     nsAString nsstr;
@@ -1235,14 +1242,21 @@
                 return E_OUTOFMEMORY;
             _wcslwr(lowercase);
             type = lowercase;
+
+            if(!wcscmp(type, L"application/xhtml+xml"))
+                doctype = DOCTYPE_XHTML;
+            else if(!wcscmp(type, L"image/svg+xml"))
+                doctype = DOCTYPE_SVG;
         }else
             type = generic_type;
     }
 
     nsAString_InitDepend(&nsstr, type);
-    nsres = nsIXMLHttpRequest_SlowOverrideMimeType(This->xhr.nsxhr, &nsstr);
+    nsres = nsIXMLHttpRequest_SlowOverrideMimeType(This->nsxhr, &nsstr);
     nsAString_Finish(&nsstr);
     free(lowercase);
+    if(NS_SUCCEEDED(nsres))
+        This->doctype_override = doctype;
     return map_nsresult(nsres);
 }
 
@@ -1252,7 +1266,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
 
-    return set_event_handler(&This->xhr.event_target, EVENTID_ERROR, &v);
+    return set_event_handler(&This->event_target, EVENTID_ERROR, &v);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_get_onerror(IWineXMLHttpRequestPrivate *iface, VARIANT *p)
@@ -1261,7 +1275,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return get_event_handler(&This->xhr.event_target, EVENTID_ERROR, p);
+    return get_event_handler(&This->event_target, EVENTID_ERROR, p);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_put_onabort(IWineXMLHttpRequestPrivate *iface, VARIANT v)
@@ -1270,7 +1284,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
 
-    return set_event_handler(&This->xhr.event_target, EVENTID_ABORT, &v);
+    return set_event_handler(&This->event_target, EVENTID_ABORT, &v);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_get_onabort(IWineXMLHttpRequestPrivate *iface, VARIANT *p)
@@ -1279,7 +1293,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return get_event_handler(&This->xhr.event_target, EVENTID_ABORT, p);
+    return get_event_handler(&This->event_target, EVENTID_ABORT, p);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_put_onprogress(IWineXMLHttpRequestPrivate *iface, VARIANT v)
@@ -1288,7 +1302,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
 
-    return set_event_handler(&This->xhr.event_target, EVENTID_PROGRESS, &v);
+    return set_event_handler(&This->event_target, EVENTID_PROGRESS, &v);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_get_onprogress(IWineXMLHttpRequestPrivate *iface, VARIANT *p)
@@ -1297,7 +1311,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return get_event_handler(&This->xhr.event_target, EVENTID_PROGRESS, p);
+    return get_event_handler(&This->event_target, EVENTID_PROGRESS, p);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_put_onloadstart(IWineXMLHttpRequestPrivate *iface, VARIANT v)
@@ -1306,7 +1320,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
 
-    return set_event_handler(&This->xhr.event_target, EVENTID_LOADSTART, &v);
+    return set_event_handler(&This->event_target, EVENTID_LOADSTART, &v);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_get_onloadstart(IWineXMLHttpRequestPrivate *iface, VARIANT *p)
@@ -1315,7 +1329,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return get_event_handler(&This->xhr.event_target, EVENTID_LOADSTART, p);
+    return get_event_handler(&This->event_target, EVENTID_LOADSTART, p);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_put_onloadend(IWineXMLHttpRequestPrivate *iface, VARIANT v)
@@ -1324,7 +1338,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
 
-    return set_event_handler(&This->xhr.event_target, EVENTID_LOADEND, &v);
+    return set_event_handler(&This->event_target, EVENTID_LOADEND, &v);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_get_onloadend(IWineXMLHttpRequestPrivate *iface, VARIANT *p)
@@ -1333,7 +1347,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return get_event_handler(&This->xhr.event_target, EVENTID_LOADEND, p);
+    return get_event_handler(&This->event_target, EVENTID_LOADEND, p);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_put_onload(IWineXMLHttpRequestPrivate *iface, VARIANT v)
@@ -1342,7 +1356,7 @@
 
     TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
 
-    return set_event_handler(&This->xhr.event_target, EVENTID_LOAD, &v);
+    return set_event_handler(&This->event_target, EVENTID_LOAD, &v);
 }
 
 static HRESULT WINAPI HTMLXMLHttpRequest_private_get_onload(IWineXMLHttpRequestPrivate *iface, VARIANT *p)
@@ -1351,7 +1365,7 @@
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    return get_event_handler(&This->xhr.event_target, EVENTID_LOAD, p);
+    return get_event_handler(&This->event_target, EVENTID_LOAD, p);
 }
 
 static const IWineXMLHttpRequestPrivateVtbl WineXMLHttpRequestPrivateVtbl = {
@@ -1430,12 +1444,7 @@
 
 static inline HTMLXMLHttpRequest *impl_from_DispatchEx(DispatchEx *iface)
 {
-    return CONTAINING_RECORD(iface, HTMLXMLHttpRequest, xhr.event_target.dispex);
-}
-
-static inline struct xhr *xhr_from_DispatchEx(DispatchEx *iface)
-{
-    return CONTAINING_RECORD(iface, struct xhr, event_target.dispex);
+    return CONTAINING_RECORD(iface, HTMLXMLHttpRequest, event_target.dispex);
 }
 
 static void *HTMLXMLHttpRequest_query_interface(DispatchEx *dispex, REFIID riid)
@@ -1453,65 +1462,65 @@
     if(IsEqualGUID(&IID_IProvideClassInfo2, riid))
         return &This->IProvideClassInfo2_iface;
 
-    return EventTarget_query_interface(&This->xhr.event_target, riid);
-}
-
-static void xhr_traverse(DispatchEx *dispex, nsCycleCollectionTraversalCallback *cb)
-{
-    struct xhr *xhr = xhr_from_DispatchEx(dispex);
-    if(xhr->window)
-        note_cc_edge((nsISupports*)&xhr->window->base.IHTMLWindow2_iface, "window", cb);
-    if(xhr->pending_progress_event)
-        note_cc_edge((nsISupports*)&xhr->pending_progress_event->IDOMEvent_iface, "pending_progress_event", cb);
-    if(xhr->response_obj)
-        note_cc_edge((nsISupports*)xhr->response_obj, "response_obj", cb);
-    if(xhr->nsxhr)
-        note_cc_edge((nsISupports*)xhr->nsxhr, "nsxhr", cb);
-    traverse_event_target(&xhr->event_target, cb);
-}
-
-static void xhr_unlink(DispatchEx *dispex)
-{
-    struct xhr *xhr = xhr_from_DispatchEx(dispex);
-    if(xhr->event_listener) {
-        XMLHttpReqEventListener *event_listener = xhr->event_listener;
-        xhr->event_listener = NULL;
+    return EventTarget_query_interface(&This->event_target, riid);
+}
+
+static void HTMLXMLHttpRequest_traverse(DispatchEx *dispex, nsCycleCollectionTraversalCallback *cb)
+{
+    HTMLXMLHttpRequest *This = impl_from_DispatchEx(dispex);
+    if(This->window)
+        note_cc_edge((nsISupports*)&This->window->base.IHTMLWindow2_iface, "window", cb);
+    if(This->pending_progress_event)
+        note_cc_edge((nsISupports*)&This->pending_progress_event->IDOMEvent_iface, "pending_progress_event", cb);
+    if(This->response_obj)
+        note_cc_edge((nsISupports*)This->response_obj, "response_obj", cb);
+    if(This->nsxhr)
+        note_cc_edge((nsISupports*)This->nsxhr, "nsxhr", cb);
+    traverse_event_target(&This->event_target, cb);
+}
+
+static void HTMLXMLHttpRequest_unlink(DispatchEx *dispex)
+{
+    HTMLXMLHttpRequest *This = impl_from_DispatchEx(dispex);
+    if(This->event_listener) {
+        XMLHttpReqEventListener *event_listener = This->event_listener;
+        This->event_listener = NULL;
         detach_xhr_event_listener(event_listener);
     }
-    if(xhr->window) {
-        HTMLInnerWindow *window = xhr->window;
-        xhr->window = NULL;
+    if(This->window) {
+        HTMLInnerWindow *window = This->window;
+        This->window = NULL;
         IHTMLWindow2_Release(&window->base.IHTMLWindow2_iface);
     }
-    if(xhr->pending_progress_event) {
-        DOMEvent *pending_progress_event = xhr->pending_progress_event;
-        xhr->pending_progress_event = NULL;
+    if(This->pending_progress_event) {
+        DOMEvent *pending_progress_event = This->pending_progress_event;
+        This->pending_progress_event = NULL;
         IDOMEvent_Release(&pending_progress_event->IDOMEvent_iface);
     }
-    unlink_ref(&xhr->response_obj);
-    unlink_ref(&xhr->nsxhr);
-    release_event_target(&xhr->event_target);
-}
-
-static void xhr_destructor(DispatchEx *dispex)
-{
-    struct xhr *xhr = xhr_from_DispatchEx(dispex);
-    free(xhr);
-}
-
-static void xhr_last_release(DispatchEx *dispex)
-{
-    struct xhr *xhr = xhr_from_DispatchEx(dispex);
-    remove_target_tasks(xhr->task_magic);
-}
-
-static nsISupports *xhr_get_gecko_target(DispatchEx *dispex)
-{
-    struct xhr *xhr = xhr_from_DispatchEx(dispex);
-    return (nsISupports*)xhr->nsxhr;
-}
-
-static void xhr_bind_event(DispatchEx *dispex, eventid_t eid)
+    unlink_ref(&This->response_obj);
+    unlink_ref(&This->nsxhr);
+    release_event_target(&This->event_target);
+}
+
+static void HTMLXMLHttpRequest_destructor(DispatchEx *dispex)
+{
+    HTMLXMLHttpRequest *This = impl_from_DispatchEx(dispex);
+    free(This);
+}
+
+static void HTMLXMLHttpRequest_last_release(DispatchEx *dispex)
+{
+    HTMLXMLHttpRequest *This = impl_from_DispatchEx(dispex);
+    remove_target_tasks(This->task_magic);
+}
+
+static nsISupports *HTMLXMLHttpRequest_get_gecko_target(DispatchEx *dispex)
+{
+    HTMLXMLHttpRequest *This = impl_from_DispatchEx(dispex);
+    return (nsISupports*)This->nsxhr;
+}
+
+static void HTMLXMLHttpRequest_bind_event(DispatchEx *dispex, eventid_t eid)
 {
     /* Do nothing. To be able to track state and queue events manually, when blocked
      * by sync XHRs in their send() event loop, we always register the handlers. */
@@ -1550,30 +1559,44 @@
 static const event_target_vtbl_t HTMLXMLHttpRequest_event_target_vtbl = {
     {
         .query_interface     = HTMLXMLHttpRequest_query_interface,
-        .destructor          = xhr_destructor,
-        .traverse            = xhr_traverse,
-        .unlink              = xhr_unlink,
-        .last_release        = xhr_last_release
+        .destructor          = HTMLXMLHttpRequest_destructor,
+        .traverse            = HTMLXMLHttpRequest_traverse,
+        .unlink              = HTMLXMLHttpRequest_unlink,
+        .last_release        = HTMLXMLHttpRequest_last_release
     },
-    .get_gecko_target        = xhr_get_gecko_target,
-    .bind_event              = xhr_bind_event
+    .get_gecko_target        = HTMLXMLHttpRequest_get_gecko_target,
+    .bind_event              = HTMLXMLHttpRequest_bind_event
 };
 
+static const tid_t HTMLXMLHttpRequest_iface_tids[] = {
+    IHTMLXMLHttpRequest2_tid,
+    0
+};
+dispex_static_data_t XMLHttpRequest_dispex = {
+    .id               = PROT_XMLHttpRequest,
+    .init_constructor = HTMLXMLHttpRequestFactory_Create,
+    .vtbl             = &HTMLXMLHttpRequest_event_target_vtbl.dispex_vtbl,
+    .disp_tid         = DispHTMLXMLHttpRequest_tid,
+    .iface_tids       = HTMLXMLHttpRequest_iface_tids,
+    .init_info        = HTMLXMLHttpRequest_init_dispex_info,
+};
+
+
 /* IHTMLXMLHttpRequestFactory */
-static inline struct constructor *impl_from_IHTMLXMLHttpRequestFactory(IHTMLXMLHttpRequestFactory *iface)
-{
-    return CONTAINING_RECORD(iface, struct constructor, iface);
+static inline HTMLXMLHttpRequestFactory *impl_from_IHTMLXMLHttpRequestFactory(IHTMLXMLHttpRequestFactory *iface)
+{
+    return CONTAINING_RECORD(iface, HTMLXMLHttpRequestFactory, IHTMLXMLHttpRequestFactory_iface);
 }
 
 DISPEX_IDISPATCH_IMPL(HTMLXMLHttpRequestFactory, IHTMLXMLHttpRequestFactory,
                       impl_from_IHTMLXMLHttpRequestFactory(iface)->dispex)
 
-static HRESULT alloc_xhr(HTMLInnerWindow *window, dispex_static_data_t *dispex, unsigned size, void **p)
+static HRESULT create_xhr(HTMLInnerWindow *window, dispex_static_data_t *dispex, HTMLXMLHttpRequest **p)
 {
     XMLHttpReqEventListener *event_listener;
     nsIDOMEventTarget *nstarget;
     nsIXMLHttpRequest *nsxhr;
-    struct xhr *ret;
+    HTMLXMLHttpRequest *ret;
     nsresult nsres;
     unsigned i;
 
@@ -1581,7 +1604,7 @@
     if(!nsxhr)
         return E_FAIL;
 
-    ret = calloc(1, size);
+    ret = calloc(1, sizeof(*ret));
     if(!ret) {
         nsIXMLHttpRequest_Release(nsxhr);
         return E_OUTOFMEMORY;
@@ -1596,9 +1619,14 @@
 
     ret->nsxhr = nsxhr;
     ret->window = window;
+    ret->doctype_override = DOCTYPE_INVALID;
     ret->task_magic = get_task_target_magic();
     IHTMLWindow2_AddRef(&window->base.IHTMLWindow2_iface);
 
+    ret->IHTMLXMLHttpRequest_iface.lpVtbl = &HTMLXMLHttpRequestVtbl;
+    ret->IHTMLXMLHttpRequest2_iface.lpVtbl = &HTMLXMLHttpRequest2Vtbl;
+    ret->IWineXMLHttpRequestPrivate_iface.lpVtbl = &WineXMLHttpRequestPrivateVtbl;
+    ret->IProvideClassInfo2_iface.lpVtbl = &ProvideClassInfo2Vtbl;
     init_event_target(&ret->event_target, dispex, window);
 
     /* Always register the handlers because we need them to track state */
@@ -1619,7 +1647,7 @@
         nsAString_Finish(&type_str);
         if(NS_FAILED(nsres)) {
             WARN("AddEventListener(%s) failed: %08lx\n", debugstr_w(name), nsres);
-            IWineJSDispatchHost_Release(&ret->event_target.dispex.IWineJSDispatchHost_iface);
+            IHTMLXMLHttpRequest_Release(&ret->IHTMLXMLHttpRequest_iface);
             return map_nsresult(nsres);
         }
     }
@@ -1629,31 +1657,15 @@
     return S_OK;
 }
 
-static HRESULT create_xhr(HTMLInnerWindow *window, HTMLXMLHttpRequest **p)
-{
-    HTMLXMLHttpRequest *ret;
-    HRESULT hres;
-
-    hres = alloc_xhr(window, &XMLHttpRequest_dispex, sizeof(*ret), (void**)&ret);
-    if(SUCCEEDED(hres)) {
-        ret->IHTMLXMLHttpRequest_iface.lpVtbl = &HTMLXMLHttpRequestVtbl;
-        ret->IHTMLXMLHttpRequest2_iface.lpVtbl = &HTMLXMLHttpRequest2Vtbl;
-        ret->IWineXMLHttpRequestPrivate_iface.lpVtbl = &WineXMLHttpRequestPrivateVtbl;
-        ret->IProvideClassInfo2_iface.lpVtbl = &ProvideClassInfo2Vtbl;
-        *p = ret;
-    }
-    return hres;
-}
-
 static HRESULT WINAPI HTMLXMLHttpRequestFactory_create(IHTMLXMLHttpRequestFactory *iface, IHTMLXMLHttpRequest **p)
 {
-    struct constructor *This = impl_from_IHTMLXMLHttpRequestFactory(iface);
+    HTMLXMLHttpRequestFactory *This = impl_from_IHTMLXMLHttpRequestFactory(iface);
     HTMLXMLHttpRequest *xhr;
     HRESULT hres;
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    hres = create_xhr(This->window, &xhr);
+    hres = create_xhr(This->window, &XMLHttpRequest_dispex, &xhr);
     if(SUCCEEDED(hres))
         *p = &xhr->IHTMLXMLHttpRequest_iface;
     return hres;
@@ -1670,20 +1682,50 @@
     HTMLXMLHttpRequestFactory_create
 };
 
+static inline HTMLXMLHttpRequestFactory *factory_from_DispatchEx(DispatchEx *iface)
+{
+    return CONTAINING_RECORD(iface, HTMLXMLHttpRequestFactory, dispex);
+}
+
 static void *HTMLXMLHttpRequestFactory_query_interface(DispatchEx *dispex, REFIID riid)
 {
-    struct constructor *This = constructor_from_DispatchEx(dispex);
+    HTMLXMLHttpRequestFactory *This = factory_from_DispatchEx(dispex);
 
     if(IsEqualGUID(&IID_IHTMLXMLHttpRequestFactory, riid))
-        return &This->iface;
+        return &This->IHTMLXMLHttpRequestFactory_iface;
 
     return NULL;
+}
+
+static void HTMLXMLHttpRequestFactory_traverse(DispatchEx *dispex, nsCycleCollectionTraversalCallback *cb)
+{
+    HTMLXMLHttpRequestFactory *This = factory_from_DispatchEx(dispex);
+
+    if(This->window)
+        note_cc_edge((nsISupports*)&This->window->base.IHTMLWindow2_iface, "window", cb);
+}
+
+static void HTMLXMLHttpRequestFactory_unlink(DispatchEx *dispex)
+{
+    HTMLXMLHttpRequestFactory *This = factory_from_DispatchEx(dispex);
+
+    if(This->window) {
+        HTMLInnerWindow *window = This->window;
+        This->window = NULL;
+        IHTMLWindow2_Release(&window->base.IHTMLWindow2_iface);
+    }
+}
+
+static void HTMLXMLHttpRequestFactory_destructor(DispatchEx *dispex)
+{
+    HTMLXMLHttpRequestFactory *This = factory_from_DispatchEx(dispex);
+    free(This);
 }
 
 static HRESULT HTMLXMLHttpRequestFactory_value(DispatchEx *iface, LCID lcid, WORD flags, DISPPARAMS *params,
         VARIANT *res, EXCEPINFO *ei, IServiceProvider *caller)
 {
-    struct constructor *This = constructor_from_DispatchEx(iface);
+    HTMLXMLHttpRequestFactory *This = factory_from_DispatchEx(iface);
     HTMLXMLHttpRequest *xhr;
     HRESULT hres;
 
@@ -1694,7 +1736,7 @@
         return E_NOTIMPL;
     }
 
-    hres = create_xhr(This->window, &xhr);
+    hres = create_xhr(This->window, &XMLHttpRequest_dispex, &xhr);
     if(FAILED(hres))
         return hres;
 
@@ -1705,9 +1747,9 @@
 
 static const dispex_static_data_vtbl_t HTMLXMLHttpRequestFactory_dispex_vtbl = {
     .query_interface  = HTMLXMLHttpRequestFactory_query_interface,
-    .destructor       = constructor_destructor,
-    .traverse         = constructor_traverse,
-    .unlink           = constructor_unlink,
+    .destructor       = HTMLXMLHttpRequestFactory_destructor,
+    .traverse         = HTMLXMLHttpRequestFactory_traverse,
+    .unlink           = HTMLXMLHttpRequestFactory_unlink,
     .value            = HTMLXMLHttpRequestFactory_value
 };
 
@@ -1716,95 +1758,63 @@
     0
 };
 static dispex_static_data_t HTMLXMLHttpRequestFactory_dispex = {
-    .name           = "XMLHttpRequest",
-    .constructor_id = OBJID_XMLHttpRequest,
+    .name           = "Function",
+    .constructor_id = PROT_XMLHttpRequest,
     .vtbl           = &HTMLXMLHttpRequestFactory_dispex_vtbl,
     .disp_tid       = IHTMLXMLHttpRequestFactory_tid,
     .iface_tids     = HTMLXMLHttpRequestFactory_iface_tids,
 };
 
-static HRESULT init_constructor(struct constructor *constr, dispex_static_data_t *dispex, const void *vtbl)
-{
-    struct constructor *create;
-    HRESULT hres;
-
-    constr->iface.lpVtbl = vtbl;
-    init_dispatch(&constr->dispex, dispex, constr->window, dispex_compat_mode(&constr->window->event_target.dispex));
-
-    if(!constr->window->jscript)
-        return S_OK;
-
-    if(!(create = malloc(sizeof(*create))))
+HRESULT HTMLXMLHttpRequestFactory_Create(HTMLInnerWindow* window, DispatchEx **ret_ptr)
+{
+    HTMLXMLHttpRequestFactory *ret;
+
+    ret = malloc(sizeof(*ret));
+    if(!ret)
         return E_OUTOFMEMORY;
-    create->iface = constr->iface;
-    create->window = constr->window;
-    IHTMLWindow2_AddRef(&constr->window->base.IHTMLWindow2_iface);
-
-    init_dispatch_from_desc(&create->dispex, constr->dispex.info, NULL, NULL);
-
-    hres = IWineJScript_InitHostConstructor(create->window->jscript, &create->dispex.IWineJSDispatchHost_iface,
-                                            L"create", &create->dispex.jsdisp);
-    if(SUCCEEDED(hres)) {
-        VARIANT v;
-        V_VT(&v) = VT_DISPATCH;
-        V_DISPATCH(&v) = (IDispatch*)&create->dispex.IWineJSDispatchHost_iface;
-        hres = IWineJSDispatch_DefineProperty(constr->dispex.jsdisp, L"create", PROPF_WRITABLE | PROPF_CONFIGURABLE, &v);
-    }
-
-    IWineJSDispatchHost_Release(&create->dispex.IWineJSDispatchHost_iface);
-    return hres;
-}
-
-static HRESULT HTMLXMLHttpRequestFactory_init(struct constructor *constr)
-{
-    return init_constructor(constr, &HTMLXMLHttpRequestFactory_dispex, &HTMLXMLHttpRequestFactoryVtbl);
-}
-
-static const tid_t HTMLXMLHttpRequest_iface_tids[] = {
-    IHTMLXMLHttpRequest2_tid,
-    0
-};
-dispex_static_data_t XMLHttpRequest_dispex = {
-    .id               = OBJID_XMLHttpRequest,
-    .init_constructor = HTMLXMLHttpRequestFactory_init,
-    .vtbl             = &HTMLXMLHttpRequest_event_target_vtbl.dispex_vtbl,
-    .disp_tid         = DispHTMLXMLHttpRequest_tid,
-    .iface_tids       = HTMLXMLHttpRequest_iface_tids,
-    .init_info        = HTMLXMLHttpRequest_init_dispex_info,
-};
-
+
+    ret->IHTMLXMLHttpRequestFactory_iface.lpVtbl = &HTMLXMLHttpRequestFactoryVtbl;
+    ret->window = window;
+    IHTMLWindow2_AddRef(&window->base.IHTMLWindow2_iface);
+
+    init_dispatch(&ret->dispex, &HTMLXMLHttpRequestFactory_dispex, window,
+                  dispex_compat_mode(&window->event_target.dispex));
+
+    *ret_ptr = &ret->dispex;
+    return S_OK;
+}
 
 
 /* IHTMLXDomainRequest */
-static inline HTMLXDomainRequest *impl_from_IHTMLXDomainRequest(IHTMLXDomainRequest *iface)
-{
-    return CONTAINING_RECORD(iface, HTMLXDomainRequest, IHTMLXDomainRequest_iface);
+static inline HTMLXMLHttpRequest *impl_from_IHTMLXDomainRequest(IHTMLXDomainRequest *iface)
+{
+    return CONTAINING_RECORD(iface, HTMLXMLHttpRequest, IHTMLXDomainRequest_iface);
 }
 
 DISPEX_IDISPATCH_IMPL(HTMLXDomainRequest, IHTMLXDomainRequest,
-                      impl_from_IHTMLXDomainRequest(iface)->xhr.event_target.dispex)
+                      impl_from_IHTMLXDomainRequest(iface)->event_target.dispex)
 
 static HRESULT WINAPI HTMLXDomainRequest_get_responseText(IHTMLXDomainRequest *iface, BSTR *p)
 {
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    return get_response_text(&This->xhr, p);
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    return HTMLXMLHttpRequest_get_responseText(&This->IHTMLXMLHttpRequest_iface, p);
 }
 
 static HRESULT WINAPI HTMLXDomainRequest_put_timeout(IHTMLXDomainRequest *iface, LONG v)
 {
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
 
     TRACE("(%p)->(%ld)\n", This, v);
 
     if(v < 0)
         return E_INVALIDARG;
-    return map_nsresult(nsIXMLHttpRequest_SetTimeout(This->xhr.nsxhr, v));
+    return map_nsresult(nsIXMLHttpRequest_SetTimeout(This->nsxhr, v));
 }
 
 static HRESULT WINAPI HTMLXDomainRequest_get_timeout(IHTMLXDomainRequest *iface, LONG *p)
 {
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
     nsresult nsres;
     UINT32 timeout;
 
@@ -1813,183 +1823,185 @@
     if(!p)
         return E_INVALIDARG;
 
-    nsres = nsIXMLHttpRequest_GetTimeout(This->xhr.nsxhr, &timeout);
+    nsres = nsIXMLHttpRequest_GetTimeout(This->nsxhr, &timeout);
     *p = timeout ? timeout : -1;
     return map_nsresult(nsres);
 }
 
 static HRESULT WINAPI HTMLXDomainRequest_get_contentType(IHTMLXDomainRequest *iface, BSTR *p)
 {
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-    const char *content_type = "text/plain";
-    nsIChannel *nschannel;
-    HRESULT hres = S_OK;
-    nsACString nsstr;
-    nsresult nsres;
-    int len;
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    if(!p)
-        return E_POINTER;
-
-    if(This->xhr.ready_state < READYSTATE_LOADED) {
-        *p = NULL;
-        return S_OK;
-    }
-
-    nsACString_Init(&nsstr, NULL);
-
-    nsres = nsIXMLHttpRequest_GetChannel(This->xhr.nsxhr, &nschannel);
-    if(NS_SUCCEEDED(nsres)) {
-        nsres = nsIChannel_GetContentType(nschannel, &nsstr);
-        if(NS_SUCCEEDED(nsres))
-            nsACString_GetData(&nsstr, &content_type);
-        nsIChannel_Release(nschannel);
-    }
-
-    len = MultiByteToWideChar(CP_UTF8, 0, content_type, -1, NULL, 0);
-    if(!(*p = SysAllocStringLen(NULL, len - 1)))
-        hres = E_OUTOFMEMORY;
-    else
-        MultiByteToWideChar(CP_UTF8, 0, content_type, -1, *p, len);
-
-    nsACString_Finish(&nsstr);
-    return hres;
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_put_onprogress(IHTMLXDomainRequest *iface, VARIANT v)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
-
-    return set_event_handler(&This->xhr.event_target, EVENTID_PROGRESS, &v);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_get_onprogress(IHTMLXDomainRequest *iface, VARIANT *p)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    return get_event_handler(&This->xhr.event_target, EVENTID_PROGRESS, p);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_put_onerror(IHTMLXDomainRequest *iface, VARIANT v)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
-
-    return set_event_handler(&This->xhr.event_target, EVENTID_ERROR, &v);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_get_onerror(IHTMLXDomainRequest *iface, VARIANT *p)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    return get_event_handler(&This->xhr.event_target, EVENTID_ERROR, p);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_put_ontimeout(IHTMLXDomainRequest *iface, VARIANT v)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
-
-    return set_event_handler(&This->xhr.event_target, EVENTID_TIMEOUT, &v);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_get_ontimeout(IHTMLXDomainRequest *iface, VARIANT *p)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    return get_event_handler(&This->xhr.event_target, EVENTID_TIMEOUT, p);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_put_onload(IHTMLXDomainRequest *iface, VARIANT v)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
-
-    return set_event_handler(&This->xhr.event_target, EVENTID_LOAD, &v);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_get_onload(IHTMLXDomainRequest *iface, VARIANT *p)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    TRACE("(%p)->(%p)\n", This, p);
-
-    return get_event_handler(&This->xhr.event_target, EVENTID_LOAD, p);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_abort(IHTMLXDomainRequest *iface)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    return xhr_abort(&This->xhr);
-}
-
-static HRESULT WINAPI HTMLXDomainRequest_open(IHTMLXDomainRequest *iface, BSTR bstrMethod, BSTR bstrUrl)
-{
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-    HTMLOuterWindow *outer_window = This->xhr.window->base.outer_window;
-    nsLoadFlags load_flags;
-    DWORD scheme, scheme2;
-    nsIChannel *nschannel;
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
     nsAString nsstr;
     nsresult nsres;
     HRESULT hres;
-    IUri *uri;
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    if(!p)
+        return E_POINTER;
+
+    if(This->ready_state < READYSTATE_LOADED) {
+        *p = NULL;
+        return S_OK;
+    }
+
+    nsAString_InitDepend(&nsstr, NULL);
+    nsres = nsIXMLHttpRequest_GetResponseText(This->nsxhr, &nsstr);
+    if(NS_SUCCEEDED(nsres)) {
+        const PRUnichar *data;
+        char text[256 * 3];
+        unsigned len;
+        WCHAR *mime;
+
+        nsAString_GetData(&nsstr, &data);
+        len = WideCharToMultiByte(CP_ACP, 0, data, wcsnlen(data, 256), text, ARRAY_SIZE(text), NULL, NULL);
+        nsAString_Finish(&nsstr);
+
+        if(len) {
+            hres = FindMimeFromData(NULL, NULL, text, len, NULL, 0, &mime, 0);
+            if(SUCCEEDED(hres)) {
+                *p = SysAllocString(mime);
+                CoTaskMemFree(mime);
+                return *p ? S_OK : E_OUTOFMEMORY;
+            }
+        }
+    }
+
+    *p = SysAllocString(L"text/plain");
+    return *p ? S_OK : E_OUTOFMEMORY;
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_put_onprogress(IHTMLXDomainRequest *iface, VARIANT v)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_event_handler(&This->event_target, EVENTID_PROGRESS, &v);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_get_onprogress(IHTMLXDomainRequest *iface, VARIANT *p)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_event_handler(&This->event_target, EVENTID_PROGRESS, p);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_put_onerror(IHTMLXDomainRequest *iface, VARIANT v)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_event_handler(&This->event_target, EVENTID_ERROR, &v);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_get_onerror(IHTMLXDomainRequest *iface, VARIANT *p)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_event_handler(&This->event_target, EVENTID_ERROR, p);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_put_ontimeout(IHTMLXDomainRequest *iface, VARIANT v)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_event_handler(&This->event_target, EVENTID_TIMEOUT, &v);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_get_ontimeout(IHTMLXDomainRequest *iface, VARIANT *p)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_event_handler(&This->event_target, EVENTID_TIMEOUT, p);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_put_onload(IHTMLXDomainRequest *iface, VARIANT v)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    TRACE("(%p)->(%s)\n", This, debugstr_variant(&v));
+
+    return set_event_handler(&This->event_target, EVENTID_LOAD, &v);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_get_onload(IHTMLXDomainRequest *iface, VARIANT *p)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    TRACE("(%p)->(%p)\n", This, p);
+
+    return get_event_handler(&This->event_target, EVENTID_LOAD, p);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_abort(IHTMLXDomainRequest *iface)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+    return HTMLXMLHttpRequest_abort(&This->IHTMLXMLHttpRequest_iface);
+}
+
+static HRESULT WINAPI HTMLXDomainRequest_open(IHTMLXDomainRequest *iface, BSTR bstrMethod, BSTR bstrUrl)
+{
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+    VARIANT vtrue, vempty;
+    nsACString str1, str2;
+    nsAString nsstr;
+    nsresult nsres;
+    HRESULT hres;
+    DWORD magic;
+    WCHAR *p;
 
     TRACE("(%p)->(%s %s)\n", This, debugstr_w(bstrMethod), debugstr_w(bstrUrl));
 
-    if(!outer_window || !outer_window->uri)
-        return E_ACCESSDENIED;
-
-    hres = create_relative_uri(outer_window, bstrUrl, &uri);
-    if(FAILED(hres))
+    if((p = wcschr(bstrUrl, ':')) && p[1] == '/' && p[2] == '/') {
+        size_t len = p - bstrUrl;
+        BSTR bstr;
+
+        /* Native only allows http and https, and the scheme must match */
+        if(len < 4 || len > 5 || wcsnicmp(bstrUrl, L"https", len) || !This->window->base.outer_window || !This->window->base.outer_window->uri)
+            return E_ACCESSDENIED;
+
+        hres = IUri_GetSchemeName(This->window->base.outer_window->uri, &bstr);
+        if(FAILED(hres))
+            return hres;
+        if(SysStringLen(bstr) != len || wcsnicmp(bstr, bstrUrl, len))
+            hres = E_ACCESSDENIED;
+        SysFreeString(bstr);
+        if(FAILED(hres))
+            return hres;
+    }
+
+    V_VT(&vtrue) = VT_BOOL;
+    V_BOOL(&vtrue) = VARIANT_TRUE;
+    V_VT(&vempty) = VT_EMPTY;
+    magic = This->magic;
+    hres = HTMLXMLHttpRequest_open(&This->IHTMLXMLHttpRequest_iface, bstrMethod, bstrUrl, vtrue, vempty, vempty);
+    if(FAILED(hres) || magic != This->magic)
         return hres;
 
-    hres = IUri_GetScheme(uri, &scheme);
-    if(hres == S_OK) {
-        if(scheme != URL_SCHEME_HTTP && scheme != URL_SCHEME_HTTPS)
-            hres = E_ACCESSDENIED;
-        else {
-            hres = IUri_GetScheme(outer_window->uri, &scheme2);
-            if(hres == S_OK && scheme != scheme2)
-                hres = E_ACCESSDENIED;
-        }
-    }
-    IUri_Release(uri);
-    if(hres != S_OK)
-        return E_ACCESSDENIED;
-
-    hres = xhr_open(&This->xhr, bstrMethod, bstrUrl, TRUE, NULL, NULL, 1);
-    if(FAILED(hres))
-        return hres;
-
     /* Prevent Gecko from parsing responseXML for no reason */
-    nsAString_InitDepend(&nsstr, L"text");
-    nsIXMLHttpRequest_SetResponseType(This->xhr.nsxhr, &nsstr);
+    nsAString_InitDepend(&nsstr, L"text/plain");
+    nsIXMLHttpRequest_SlowOverrideMimeType(This->nsxhr, &nsstr);
     nsAString_Finish(&nsstr);
 
-    /* Always sniff the content type even there's a Content-Type response header, to match native */
-    nsres = nsIXMLHttpRequest_GetChannel(This->xhr.nsxhr, &nschannel);
-    if(NS_SUCCEEDED(nsres)) {
-        nsres = nsIChannel_GetLoadFlags(nschannel, &load_flags);
-        if(NS_SUCCEEDED(nsres)) {
-            load_flags |= LOAD_CALL_CONTENT_SNIFFERS;
-            nsIChannel_SetLoadFlags(nschannel, load_flags);
-        }
-        nsIChannel_Release(nschannel);
+    /* XDomainRequest only accepts text/plain */
+    nsACString_InitDepend(&str1, "Accept");
+    nsACString_InitDepend(&str2, "text/plain");
+    nsres = nsIXMLHttpRequest_SetRequestHeader(This->nsxhr, &str1, &str2);
+    nsACString_Finish(&str1);
+    nsACString_Finish(&str2);
+    if(NS_FAILED(nsres)) {
+        ERR("nsIXMLHttpRequest_SetRequestHeader failed: %08lx\n", nsres);
+        return map_nsresult(nsres);
     }
 
     /* IE always adds Origin header, even from same origin, but Gecko doesn't allow us to alter it. */
@@ -1998,9 +2010,9 @@
 
 static HRESULT WINAPI HTMLXDomainRequest_send(IHTMLXDomainRequest *iface, VARIANT varBody)
 {
-    HTMLXDomainRequest *This = impl_from_IHTMLXDomainRequest(iface);
-
-    return xhr_send(&This->xhr, &varBody);
+    HTMLXMLHttpRequest *This = impl_from_IHTMLXDomainRequest(iface);
+
+    return HTMLXMLHttpRequest_send(&This->IHTMLXMLHttpRequest_iface, varBody);
 }
 
 static const IHTMLXDomainRequestVtbl HTMLXDomainRequestVtbl = {
@@ -2028,31 +2040,26 @@
     HTMLXDomainRequest_send
 };
 
-static inline HTMLXDomainRequest *HTMLXDomainRequest_from_DispatchEx(DispatchEx *iface)
-{
-    return CONTAINING_RECORD(iface, HTMLXDomainRequest, xhr.event_target.dispex);
-}
-
 static void *HTMLXDomainRequest_query_interface(DispatchEx *dispex, REFIID riid)
 {
-    HTMLXDomainRequest *This = HTMLXDomainRequest_from_DispatchEx(dispex);
+    HTMLXMLHttpRequest *This = impl_from_DispatchEx(dispex);
 
     if(IsEqualGUID(&IID_IHTMLXDomainRequest, riid))
         return &This->IHTMLXDomainRequest_iface;
 
-    return EventTarget_query_interface(&This->xhr.event_target, riid);
+    return EventTarget_query_interface(&This->event_target, riid);
 }
 
 static const event_target_vtbl_t HTMLXDomainRequest_event_target_vtbl = {
     {
         .query_interface     = HTMLXDomainRequest_query_interface,
-        .destructor          = xhr_destructor,
-        .traverse            = xhr_traverse,
-        .unlink              = xhr_unlink,
-        .last_release        = xhr_last_release
+        .destructor          = HTMLXMLHttpRequest_destructor,
+        .traverse            = HTMLXMLHttpRequest_traverse,
+        .unlink              = HTMLXMLHttpRequest_unlink,
+        .last_release        = HTMLXMLHttpRequest_last_release
     },
-    .get_gecko_target        = xhr_get_gecko_target,
-    .bind_event              = xhr_bind_event
+    .get_gecko_target        = HTMLXMLHttpRequest_get_gecko_target,
+    .bind_event              = HTMLXMLHttpRequest_bind_event
 };
 
 static void HTMLXDomainRequest_init_dispex_info(dispex_data_t *info, compat_mode_t compat_mode)
@@ -2060,11 +2067,20 @@
     dispex_info_add_interface(info, IHTMLXDomainRequest_tid, NULL);
 }
 
+dispex_static_data_t XDomainRequest_dispex = {
+    .id               = PROT_XDomainRequest,
+    .init_constructor = HTMLXDomainRequestFactory_Create,
+    .vtbl             = &HTMLXDomainRequest_event_target_vtbl.dispex_vtbl,
+    .disp_tid         = DispXDomainRequest_tid,
+    .init_info        = HTMLXDomainRequest_init_dispex_info,
+    .max_compat_mode  = COMPAT_MODE_IE10
+};
+
 
 /* IHTMLXDomainRequestFactory */
-static inline struct constructor *impl_from_IHTMLXDomainRequestFactory(IHTMLXDomainRequestFactory *iface)
-{
-    return CONTAINING_RECORD(iface, struct constructor, iface);
+static inline HTMLXDomainRequestFactory *impl_from_IHTMLXDomainRequestFactory(IHTMLXDomainRequestFactory *iface)
+{
+    return CONTAINING_RECORD(iface, HTMLXDomainRequestFactory, IHTMLXDomainRequestFactory_iface);
 }
 
 DISPEX_IDISPATCH_IMPL(HTMLXDomainRequestFactory, IHTMLXDomainRequestFactory,
@@ -2072,23 +2088,23 @@
 
 static HRESULT WINAPI HTMLXDomainRequestFactory_create(IHTMLXDomainRequestFactory *iface, IHTMLXDomainRequest **p)
 {
-    struct constructor *This = impl_from_IHTMLXDomainRequestFactory(iface);
-    HTMLXDomainRequest *xdr;
+    HTMLXDomainRequestFactory *This = impl_from_IHTMLXDomainRequestFactory(iface);
+    HTMLXMLHttpRequest *xhr;
     HRESULT hres;
 
     TRACE("(%p)->(%p)\n", This, p);
 
-    hres = alloc_xhr(This->window, &XDomainRequest_dispex, sizeof(*xdr), (void**)&xdr);
+    hres = create_xhr(This->window, &XDomainRequest_dispex, &xhr);
     if(FAILED(hres))
         return hres;
 
-    xdr->IHTMLXDomainRequest_iface.lpVtbl = &HTMLXDomainRequestVtbl;
-
-    *p = &xdr->IHTMLXDomainRequest_iface;
+    xhr->IHTMLXDomainRequest_iface.lpVtbl = &HTMLXDomainRequestVtbl;
+
+    *p = &xhr->IHTMLXDomainRequest_iface;
     return hres;
 }
 
-static const IHTMLXDomainRequestFactoryVtbl HTMLXDomainRequestFactoryVtbl = {
+const IHTMLXDomainRequestFactoryVtbl HTMLXDomainRequestFactoryVtbl = {
     HTMLXDomainRequestFactory_QueryInterface,
     HTMLXDomainRequestFactory_AddRef,
     HTMLXDomainRequestFactory_Release,
@@ -2099,20 +2115,50 @@
     HTMLXDomainRequestFactory_create
 };
 
+static inline HTMLXDomainRequestFactory *HTMLXDomainRequestFactory_from_DispatchEx(DispatchEx *iface)
+{
+    return CONTAINING_RECORD(iface, HTMLXDomainRequestFactory, dispex);
+}
+
 static void *HTMLXDomainRequestFactory_query_interface(DispatchEx *dispex, REFIID riid)
 {
-    struct constructor *This = constructor_from_DispatchEx(dispex);
+    HTMLXDomainRequestFactory *This = HTMLXDomainRequestFactory_from_DispatchEx(dispex);
 
     if(IsEqualGUID(&IID_IHTMLXDomainRequestFactory, riid))
-        return &This->iface;
+        return &This->IHTMLXDomainRequestFactory_iface;
 
     return NULL;
+}
+
+static void HTMLXDomainRequestFactory_traverse(DispatchEx *dispex, nsCycleCollectionTraversalCallback *cb)
+{
+    HTMLXDomainRequestFactory *This = HTMLXDomainRequestFactory_from_DispatchEx(dispex);
+
+    if(This->window)
+        note_cc_edge((nsISupports*)&This->window->base.IHTMLWindow2_iface, "window", cb);
+}
+
+static void HTMLXDomainRequestFactory_unlink(DispatchEx *dispex)
+{
+    HTMLXDomainRequestFactory *This = HTMLXDomainRequestFactory_from_DispatchEx(dispex);
+
+    if(This->window) {
+        HTMLInnerWindow *window = This->window;
+        This->window = NULL;
+        IHTMLWindow2_Release(&window->base.IHTMLWindow2_iface);
+    }
+}
+
+static void HTMLXDomainRequestFactory_destructor(DispatchEx *dispex)
+{
+    HTMLXDomainRequestFactory *This = HTMLXDomainRequestFactory_from_DispatchEx(dispex);
+    free(This);
 }
 
 static HRESULT HTMLXDomainRequestFactory_value(DispatchEx *dispex, LCID lcid, WORD flags, DISPPARAMS *params,
         VARIANT *res, EXCEPINFO *ei, IServiceProvider *caller)
 {
-    struct constructor *This = constructor_from_DispatchEx(dispex);
+    HTMLXDomainRequestFactory *This = HTMLXDomainRequestFactory_from_DispatchEx(dispex);
     IHTMLXDomainRequest *xdr;
     HRESULT hres;
 
@@ -2123,7 +2169,7 @@
         return E_NOTIMPL;
     }
 
-    hres = HTMLXDomainRequestFactory_create((IHTMLXDomainRequestFactory*)&This->iface, &xdr);
+    hres = HTMLXDomainRequestFactory_create(&This->IHTMLXDomainRequestFactory_iface, &xdr);
     if(FAILED(hres))
         return hres;
 
@@ -2134,9 +2180,9 @@
 
 static const dispex_static_data_vtbl_t HTMLXDomainRequestFactory_dispex_vtbl = {
     .query_interface  = HTMLXDomainRequestFactory_query_interface,
-    .destructor       = constructor_destructor,
-    .traverse         = constructor_traverse,
-    .unlink           = constructor_unlink,
+    .destructor       = HTMLXDomainRequestFactory_destructor,
+    .traverse         = HTMLXDomainRequestFactory_traverse,
+    .unlink           = HTMLXDomainRequestFactory_unlink,
     .value            = HTMLXDomainRequestFactory_value,
 };
 
@@ -2145,23 +2191,26 @@
     0
 };
 static dispex_static_data_t HTMLXDomainRequestFactory_dispex = {
-    .name            = "XDomainRequest",
-    .constructor_id  = OBJID_XDomainRequest,
+    .name            = "Function",
+    .constructor_id  = PROT_XDomainRequest,
     .vtbl            = &HTMLXDomainRequestFactory_dispex_vtbl,
     .disp_tid        = IHTMLXDomainRequestFactory_tid,
     .iface_tids      = HTMLXDomainRequestFactory_iface_tids,
 };
 
-static HRESULT HTMLXDomainRequestFactory_init(struct constructor *constr)
-{
-    return init_constructor(constr, &HTMLXDomainRequestFactory_dispex, &HTMLXDomainRequestFactoryVtbl);
-}
-
-dispex_static_data_t XDomainRequest_dispex = {
-    .id               = OBJID_XDomainRequest,
-    .init_constructor = &HTMLXDomainRequestFactory_init,
-    .vtbl             = &HTMLXDomainRequest_event_target_vtbl.dispex_vtbl,
-    .disp_tid         = DispXDomainRequest_tid,
-    .init_info        = HTMLXDomainRequest_init_dispex_info,
-    .max_compat_mode  = COMPAT_MODE_IE10
-};
+HRESULT HTMLXDomainRequestFactory_Create(HTMLInnerWindow* window, DispatchEx **ret)
+{
+    HTMLXDomainRequestFactory *xdr;
+
+    if(!(xdr = malloc(sizeof(*xdr))))
+        return E_OUTOFMEMORY;
+
+    xdr->IHTMLXDomainRequestFactory_iface.lpVtbl = &HTMLXDomainRequestFactoryVtbl;
+    xdr->window = window;
+    IHTMLWindow2_AddRef(&window->base.IHTMLWindow2_iface);
+
+    init_dispatch(&xdr->dispex, &HTMLXDomainRequestFactory_dispex, window, dispex_compat_mode(&window->event_target.dispex));
+
+    *ret = &xdr->dispex;
+    return S_OK;
+}
