diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -229,7 +229,6 @@
     struct file_identity    id;      /* directory file identity */
     struct dir_data_names  *names;   /* directory file names */
     struct dir_data_buffer *buffer;  /* head of data buffers list */
-    UNICODE_STRING          mask;    /* the mask used when creating the cache entry */
 };
 
 static const unsigned int dir_data_buffer_initial_size = 4096;
@@ -241,11 +240,6 @@
 
 static BOOL show_dot_files;
 static mode_t start_umask;
-
-static const WCHAR nt_prefixW[] = {'\\','?','?','\\'};
-static const WCHAR dos_prefixW[] = {'\\','?','?','\\','A',':','\\'};
-static const WCHAR unc_prefixW[] = {'\\','?','?','\\','U','N','C','\\'};
-static const WCHAR unix_prefixW[] = {'\\','?','?','\\','u','n','i','x'};
 
 /* at some point we may want to allow Winelib apps to set this */
 static const BOOL is_case_sensitive = FALSE;
@@ -577,7 +571,6 @@
         free( buffer );
     }
     free( data->names );
-    free( data->mask.Buffer );
     free( data );
 }
 
@@ -590,19 +583,21 @@
     char name[1];
 };
 
-static NTSTATUS add_dir_to_queue( struct list *queue, const char *name )
+static struct list dir_queue = LIST_INIT( dir_queue );
+
+static NTSTATUS add_dir_to_queue( const char *name )
 {
     int len = strlen( name ) + 1;
     struct dir_name *dir = malloc( offsetof( struct dir_name, name[len] ));
     if (!dir) return STATUS_NO_MEMORY;
     strcpy( dir->name, name );
-    list_add_tail( queue, &dir->entry );
+    list_add_tail( &dir_queue, &dir->entry );
     return STATUS_SUCCESS;
 }
 
-static NTSTATUS next_dir_in_queue( struct list *queue, char *name )
-{
-    struct list *head = list_head( queue );
+static NTSTATUS next_dir_in_queue( char *name )
+{
+    struct list *head = list_head( &dir_queue );
     if (head)
     {
         struct dir_name *dir = LIST_ENTRY( head, struct dir_name, entry );
@@ -614,11 +609,11 @@
     return STATUS_OBJECT_NAME_NOT_FOUND;
 }
 
-static void flush_dir_queue( struct list *queue )
+static void flush_dir_queue(void)
 {
     struct list *head;
 
-    while ((head = list_head( queue )))
+    while ((head = list_head( &dir_queue )))
     {
         struct dir_name *dir = LIST_ENTRY( head, struct dir_name, entry );
         list_remove( &dir->entry );
@@ -1303,6 +1298,23 @@
 #endif
     return TRUE;
 
+#elif defined(__NetBSD__)
+    struct statvfs stfs;
+    int fd;
+
+    if ((fd = openat( root_fd, dir, O_RDONLY )) == -1) return TRUE;
+    if (fstatvfs( fd, &stfs ) == -1)
+    {
+        close( fd );
+        return TRUE;
+    }
+    close( fd );
+    /* Only assume CIOPFS is case insensitive. */
+    if (strcmp( stfs.f_fstypename, "fusefs" ) ||
+        strncmp( stfs.f_mntfromname, "ciopfs", 5 ))
+        return FALSE;
+    return TRUE;
+
 #elif defined(__linux__)
     BOOLEAN sens = TRUE;
     struct statfs stfs;
@@ -1736,9 +1748,24 @@
     else return xattr_fremove( fd, SAMBA_XATTR_DOS_ATTRIB );
 }
 
+static unsigned int server_get_unix_name( HANDLE handle, char **unix_name );
+
+/* return TRUE if this is a file owned by Wine which applications should not try to mess with. */
+static BOOL is_wine_file( HANDLE handle )
+{
+    char *unix_name;
+    BOOL ret;
+
+    if (server_get_unix_name( handle, &unix_name ))
+        return FALSE;
+    ret = strstr(unix_name, "/lib/wine/" ) || strstr( unix_name, "/lib64/wine/" ) ||strstr( unix_name, "/share/wine/" );
+    free(unix_name);
+    return ret;
+}
+
 
 /* set the stat info and file attributes for a file (by file descriptor) */
-static NTSTATUS fd_set_file_info( int fd, UINT attr, BOOL force_set_xattr )
+static NTSTATUS fd_set_file_info( int fd, HANDLE handle, UINT attr, BOOL force_set_xattr )
 {
     struct stat st;
 
@@ -1752,8 +1779,16 @@
     }
     else
     {
-        /* add write permission only where we already have read permission */
-        st.st_mode |= (0600 | ((st.st_mode & 044) >> 1)) & (~start_umask);
+        if (is_wine_file( handle ))
+        {
+            TRACE("HACK: Not giving write permission to wine file!\n");
+            return STATUS_ACCESS_DENIED;
+        }
+        else
+        {
+            /* add write permission only where we already have read permission */
+            st.st_mode |= (0600 | ((st.st_mode & 044) >> 1)) & (~start_umask);
+        }
     }
     if (fchmod( fd, st.st_mode ) == -1) return errno_to_status( errno );
 
@@ -1781,6 +1816,7 @@
     *attr = 0;
     ret = lstat( path, st );
     if (ret == -1) return ret;
+    if (reparse_tag) *reparse_tag = 0;
     if (S_ISLNK( st->st_mode ))
     {
         ret = stat( path, st );
@@ -2086,6 +2122,7 @@
     case FileIdExtdBothDirectoryInformation:
         {
             FILE_ID_EXTD_BOTH_DIRECTORY_INFORMATION *info = ptr;
+            memset( &info->FileId, 0, sizeof(info->FileId) );
             *(ULONGLONG *)&info->FileId = st->st_ino;
             fill_file_info( st, attr, info, FileDirectoryInformation );
         }
@@ -2145,7 +2182,7 @@
     for (;;)
     {
         if (!(name = malloc( size ))) return STATUS_NO_MEMORY;
-        if (!(status = NtQueryObject( handle, ObjectNameInformation, name, size, &size )))
+        if (!(status = NtQueryObject( handle, ObjectNameInformation, name, size, (ULONG *)&size )))
         {
             const WCHAR *ptr = name->Name.Buffer;
             const WCHAR *end = ptr + name->Name.Length / sizeof(WCHAR);
@@ -2160,7 +2197,7 @@
             if (*name_len < info->FileNameLength) status = STATUS_BUFFER_OVERFLOW;
             else if (!info->FileNameLength) status = STATUS_INVALID_INFO_CLASS;
             else *name_len = info->FileNameLength;
-            if (info->FileNameLength) memcpy( info->FileName, ptr, *name_len );
+            memcpy( info->FileName, ptr, *name_len );
             free( name );
         }
         else
@@ -2463,30 +2500,32 @@
         break;
 
     case FileFullDirectoryInformation:
-        info->full.EaSize = 0; /* FIXME */
+        /* non-Extd classes return the reparse tag in EaSize if there is one */
+        info->full.EaSize = reparse_tag;
         info->full.FileNameLength = name_len;
         break;
 
     case FileIdFullDirectoryInformation:
-        info->id_full.EaSize = 0; /* FIXME */
+        info->id_full.EaSize = reparse_tag;
         info->id_full.FileNameLength = name_len;
         break;
 
     case FileBothDirectoryInformation:
-        info->both.EaSize = 0; /* FIXME */
+        info->both.EaSize = reparse_tag;
         info->both.ShortNameLength = wcslen( names->short_name ) * sizeof(WCHAR);
         memcpy( info->both.ShortName, names->short_name, info->both.ShortNameLength );
         info->both.FileNameLength = name_len;
         break;
 
     case FileIdBothDirectoryInformation:
-        info->id_both.EaSize = 0; /* FIXME */
+        info->id_both.EaSize = reparse_tag;
         info->id_both.ShortNameLength = wcslen( names->short_name ) * sizeof(WCHAR);
         memcpy( info->id_both.ShortName, names->short_name, info->id_both.ShortNameLength );
         info->id_both.FileNameLength = name_len;
         break;
 
     case FileIdExtdBothDirectoryInformation:
+        /* Extd classes do *not* return the reparse tag in EaSize */
         info->extd_both.EaSize = 0; /* FIXME */
         info->extd_both.ReparsePointTag = reparse_tag;
         info->extd_both.ShortNameLength = wcslen( names->short_name ) * sizeof(WCHAR);
@@ -2496,6 +2535,7 @@
 
     case FileIdGlobalTxDirectoryInformation:
         info->id_tx.TxInfoFlags = 0;
+        memset( &info->id_tx.LockingTransactionId, 0, sizeof(GUID) );
         info->id_tx.FileNameLength = name_len;
         break;
 
@@ -2580,7 +2620,7 @@
 static NTSTATUS read_directory_data_getattrlist( struct dir_data *data, const char *unix_name )
 {
     struct attrlist attrlist;
-#pragma pack(push,4)
+#include "pshpack4.h"
     struct
     {
         u_int32_t length;
@@ -2588,7 +2628,7 @@
         fsobj_type_t type;
         char name[NAME_MAX * 3 + 1];
     } buffer;
-#pragma pack(pop)
+#include "poppack.h"
 
     memset( &attrlist, 0, sizeof(attrlist) );
     attrlist.bitmapcount = ATTR_BIT_MAP_COUNT;
@@ -2728,17 +2768,6 @@
         return status;
     }
 
-    if (mask)
-    {
-        data->mask.Length = data->mask.MaximumLength = mask->Length;
-        if (!(data->mask.Buffer = malloc( mask->Length )))
-        {
-            free_dir_data( data );
-            return STATUS_NO_MEMORY;
-        }
-        memcpy(data->mask.Buffer, mask->Buffer, mask->Length);
-    }
-
     /* sort filenames, but not "." and ".." */
     i = 0;
     if (i < data->count && !strcmp( data->names[i].unix_name, "." )) i++;
@@ -2758,22 +2787,6 @@
 
     *data_ret = data;
     return data->count ? STATUS_SUCCESS : STATUS_NO_SUCH_FILE;
-}
-
-
-/***********************************************************************
- *           ustring_equal
- *
- * Simplified version of RtlEqualUnicodeString that performs only case-sensitive comparisons.
- */
-static BOOLEAN ustring_equal( const UNICODE_STRING *a, const UNICODE_STRING *b )
-{
-    USHORT length_a = (a ? a->Length : 0);
-    USHORT length_b = (b ? b->Length : 0);
-
-    if (length_a != length_b) return FALSE;
-    if (length_a == 0) return TRUE;
-    return !memcmp(a->Buffer, b->Buffer, a->Length);
 }
 
 
@@ -2783,12 +2796,11 @@
  * Retrieve the cached directory data, or initialize it if necessary.
  */
 static unsigned int get_cached_dir_data( HANDLE handle, struct dir_data **data_ret, int fd,
-                                         const UNICODE_STRING *mask, BOOLEAN restart_scan )
+                                         const UNICODE_STRING *mask )
 {
     unsigned int i;
     int entry = -1, free_entries[16];
     unsigned int status;
-    BOOLEAN fresh_handle;
 
     SERVER_START_REQ( get_directory_cache_entry )
     {
@@ -2825,28 +2837,63 @@
         dir_data_cache_size = size;
     }
 
-    fresh_handle = !dir_data_cache[entry];
-
-    if (dir_data_cache[entry] && restart_scan && mask &&
-        !ustring_equal(&dir_data_cache[entry]->mask, mask))
-    {
-        TRACE( "invalidating existing cache entry for handle %p, old mask: \"%s\", new mask: \"%s\"\n",
-               handle, debugstr_us(&(dir_data_cache[entry]->mask)), debugstr_us(mask));
-        free_dir_data( dir_data_cache[entry] );
-        dir_data_cache[entry] = NULL;
-    }
-
-    if (!dir_data_cache[entry])
-    {
-        status = init_cached_dir_data( &dir_data_cache[entry], fd, mask );
-        if (status == STATUS_NO_SUCH_FILE && !fresh_handle) status = STATUS_NO_MORE_FILES;
-    }
+    if (!dir_data_cache[entry]) status = init_cached_dir_data( &dir_data_cache[entry], fd, mask );
 
     *data_ret = dir_data_cache[entry];
-    if (restart_scan) (*data_ret)->pos = 0;
     return status;
 }
 
+static NTSTATUS server_query_directory_file( HANDLE handle, IO_STATUS_BLOCK *io, void *buffer, ULONG length,
+                                             FILE_INFORMATION_CLASS info_class, BOOLEAN single_entry,
+                                             BOOLEAN restart_scan )
+{
+    FILE_DIRECTORY_INFORMATION *info = (FILE_DIRECTORY_INFORMATION *)buffer;
+    struct directory_file_entry *entries;
+    unsigned int total_len;
+    NTSTATUS status;
+
+    if (!(entries = malloc( length ))) return STATUS_NO_MEMORY;
+
+    SERVER_START_REQ( query_directory_file )
+    {
+        req->handle = wine_server_obj_handle( handle );;
+        req->restart_scan = restart_scan;
+        wine_server_set_reply( req, entries, length );
+        status = wine_server_call( req );
+        total_len = reply->total_len;
+        if (status == STATUS_OBJECT_TYPE_MISMATCH) status = STATUS_BAD_DEVICE_TYPE;
+    }
+    SERVER_END_REQ;
+
+    if (!status && offsetof(FILE_DIRECTORY_INFORMATION, FileName) + total_len > length)
+        status = STATUS_INFO_LENGTH_MISMATCH;
+
+    io->Status = status;
+    io->Information = 0;
+    if (!status)
+    {
+        static int once;
+
+        if (info_class != FileDirectoryInformation)
+        {
+            FIXME( "Unsupprted info_class %d.\n", info_class );
+            free( entries );
+            return STATUS_NOT_SUPPORTED;
+        }
+        if (!single_entry) FIXME( "Multiple entries not supported.\n" );
+
+        memset( info, 0, sizeof(*info) );
+        if (!once++) FIXME( "Not filling attributes." );
+        info->FileNameLength = entries->name_len;
+        memcpy( info->FileName, entries + 1, entries->name_len );
+        io->Information = offsetof(FILE_DIRECTORY_INFORMATION, FileName) + total_len;
+        TRACE( "-> %s.\n", debugstr_wn(info->FileName, info->FileNameLength / 2) );
+    }
+    else TRACE( "-> status %#x.\n", (int)status );
+
+    free( entries );
+    return status;
+}
 
 /******************************************************************************
  *              NtQueryDirectoryFile   (NTDLL.@)
@@ -2863,7 +2910,7 @@
 
     TRACE("(%p %p %p %p %p %p 0x%08x 0x%08x 0x%08x %s 0x%08x\n",
           handle, event, apc_routine, apc_context, io, buffer,
-          length, info_class, single_entry, debugstr_us(mask),
+          (int)length, info_class, single_entry, debugstr_us(mask),
           restart_scan);
 
     if (event || apc_routine)
@@ -2898,7 +2945,11 @@
     if (!buffer) return STATUS_ACCESS_VIOLATION;
 
     if ((status = server_get_unix_fd( handle, FILE_LIST_DIRECTORY, &fd, &needs_close, &type, NULL )))
+    {
+        if (status == STATUS_BAD_DEVICE_TYPE)
+            return server_query_directory_file( handle, io, buffer, length, info_class, single_entry, restart_scan);
         return status;
+    }
 
     if (type != FD_TYPE_DIR)
     {
@@ -2907,16 +2958,17 @@
     }
 
     io->Information = 0;
-    if (mask && mask->Length == 0) mask = NULL;
 
     mutex_lock( &dir_mutex );
 
     cwd = open( ".", O_RDONLY );
     if (fchdir( fd ) != -1)
     {
-        if (!(status = get_cached_dir_data( handle, &data, fd, mask, restart_scan )))
+        if (!(status = get_cached_dir_data( handle, &data, fd, mask )))
         {
             union file_directory_info *last_info = NULL;
+
+            if (restart_scan) data->pos = 0;
 
             while (!status && data->pos < data->count)
             {
@@ -2927,12 +2979,12 @@
 
             if (!last_info) status = STATUS_NO_MORE_FILES;
             else if (status == STATUS_MORE_ENTRIES) status = STATUS_SUCCESS;
+
+            io->Status = status;
         }
         if (cwd == -1 || fchdir( cwd ) == -1) chdir( "/" );
     }
     else status = errno_to_status( errno );
-
-    if (status != STATUS_NO_SUCH_FILE) io->Status = status;
 
     mutex_unlock( &dir_mutex );
 
@@ -3072,6 +3124,67 @@
     return STATUS_OBJECT_NAME_NOT_FOUND;
 }
 
+/* CW-Bug-Id: #23185 Emulate Steam Input native hooks for native SDL */
+static BOOL replace_steam_input_path( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *redir )
+{
+    static const WCHAR pipe_prefixW[] =
+    {
+        '\\','?','?','\\','p','i','p','e','\\','H','I','D','#','V','I','D','_','0','4','5','E',
+        '&','P','I','D','_','0','2','8','E','&','I','G','_','0','0',
+    };
+    static const WCHAR hid_prefixW[] =
+    {
+        '\\','?','?','\\','h','i','d','#','v','i','d','_','2','8','d','e',
+        '&','p','i','d','_','1','1','f','f','&','i','g','_','0'
+    };
+    static const WCHAR hid_midW[] =
+    {
+        '#','0',
+    };
+    static const WCHAR hid_tailW[] =
+    {
+        '&','0','&','0','&','1','#','{','4','d','1','e','5','5','b','2','-','f','1','6','f','-',
+        '1','1','c','f','-','8','8','c','b','-','0','0','1','1','1','1','0','0','0','0','3','0','}'
+    };
+    UNICODE_STRING *path = attr->ObjectName;
+    const WCHAR *slot = NULL, *slot_end = NULL, *serial, *serial_end = NULL;
+    UINT len = 0;
+
+    if (!path || !path->Buffer || path->Length <= sizeof(pipe_prefixW)) return FALSE;
+    if (wcsnicmp( path->Buffer, pipe_prefixW, ARRAY_SIZE(pipe_prefixW) )) return FALSE;
+
+    serial = path->Buffer + path->Length / sizeof(WCHAR);
+    while (serial > path->Buffer && *serial != '&')
+    {
+        if (*serial == '#')
+        {
+            slot_end = serial_end;
+            serial_end = serial;
+            slot = serial_end + 1;
+        }
+        serial--;
+    }
+    if (serial == path->Buffer || *serial != '&' || !slot_end || !serial_end) return FALSE;
+
+    redir->Length = sizeof(hid_prefixW) + sizeof(hid_midW) + sizeof(hid_tailW);
+    redir->Length += (serial_end - serial + slot_end - slot) * sizeof(WCHAR);
+    redir->MaximumLength = redir->Length + sizeof(WCHAR);
+    if (!(redir->Buffer = malloc( redir->MaximumLength ))) return FALSE;
+
+    memcpy( redir->Buffer, hid_prefixW, sizeof(hid_prefixW) );
+    len += ARRAY_SIZE(hid_prefixW);
+    memcpy( redir->Buffer + len, slot, (slot_end - slot) * sizeof(WCHAR) );
+    len += slot_end - slot;
+    memcpy( redir->Buffer + len, hid_midW, sizeof(hid_midW) );
+    len += ARRAY_SIZE(hid_midW);
+    memcpy( redir->Buffer + len, serial, (serial_end - serial) * sizeof(WCHAR) );
+    len += serial_end - serial;
+    memcpy( redir->Buffer + len, hid_tailW, sizeof(hid_tailW) );
+
+    TRACE( "HACK: %s -> %s\n", debugstr_us(attr->ObjectName), debugstr_us(redir) );
+    attr->ObjectName = redir;
+    return TRUE;
+}
 
 #ifndef _WIN64
 
@@ -3167,61 +3280,77 @@
 /***********************************************************************
  *           get_redirect
  */
-static void get_redirect( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *redir )
+BOOL get_redirect( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *redir )
 {
     const WCHAR *name = attr->ObjectName->Buffer;
     unsigned int i, prefix_len = 0, len = attr->ObjectName->Length / sizeof(WCHAR);
 
-    if (!NtCurrentTeb64()) return;
+    redir->Buffer = NULL;
+
+    /* CW-Bug-Id: #23185 Emulate Steam Input native hooks for native SDL */
+    if (replace_steam_input_path( attr, redir )) return TRUE;
+
+    if (!NtCurrentTeb64()) return FALSE;
+    if (!len) return FALSE;
 
     if (!attr->RootDirectory)
     {
         prefix_len = wcslen( windirW );
-        if (len < prefix_len || wcsnicmp( name, windirW, prefix_len )) return;
+        if (len < prefix_len || wcsnicmp( name, windirW, prefix_len )) return FALSE;
     }
     else
     {
         int fd, needs_close;
         struct stat st;
 
-        if (!len) return;
-        if (server_get_unix_fd( attr->RootDirectory, 0, &fd, &needs_close, NULL, NULL )) return;
+        if (server_get_unix_fd( attr->RootDirectory, 0, &fd, &needs_close, NULL, NULL )) return FALSE;
         fstat( fd, &st );
         if (needs_close) close( fd );
         if (!is_same_file( &windir, &st ))
         {
-            if (!is_same_file( &sysdir, &st )) return;
-            if (NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR]) return;
-            if (name[0] == '\\') return;
+            if (!is_same_file( &sysdir, &st )) return FALSE;
+            if (NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR]) return FALSE;
+            if (name[0] == '\\') return FALSE;
 
             /* only check for paths that should NOT be redirected */
             for (i = 0; i < ARRAY_SIZE( no_redirect ); i++)
-                if (starts_with_path( name, len, no_redirect[i] + 9 /* "system32\\" */)) return;
+                if (starts_with_path( name, len, no_redirect[i] + 9 /* "system32\\" */)) return FALSE;
 
             /* redirect everything else */
             redir->Length = sizeof(syswow64dirW) + len * sizeof(WCHAR);
             redir->MaximumLength = redir->Length + sizeof(WCHAR);
-            if (!(redir->Buffer = malloc( redir->MaximumLength ))) return;
+            if (!(redir->Buffer = malloc( redir->MaximumLength ))) return FALSE;
             memcpy( redir->Buffer, syswow64dirW, sizeof(syswow64dirW) );
             memcpy( redir->Buffer + ARRAY_SIZE(syswow64dirW), name, len * sizeof(WCHAR) );
             redir->Buffer[redir->Length / sizeof(WCHAR)] = 0;
             attr->RootDirectory = 0;
             attr->ObjectName = redir;
-            return;
+            return TRUE;
         }
     }
 
     /* sysnative is redirected even when redirection is disabled */
 
-    if (replace_path( attr, redir, prefix_len, sysnativeW, system32W )) return;
-
-    if (NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR]) return;
+    if (replace_path( attr, redir, prefix_len, sysnativeW, system32W )) return TRUE;
+
+    if (NtCurrentTeb64()->TlsSlots[WOW64_TLS_FILESYSREDIR]) return FALSE;
 
     for (i = 0; i < ARRAY_SIZE( no_redirect ); i++)
-        if (starts_with_path( name + prefix_len, len - prefix_len, no_redirect[i] )) return;
-
-    if (replace_path( attr, redir, prefix_len, system32W, syswow64W )) return;
-    if (replace_path( attr, redir, prefix_len, regeditW, syswow64_regeditW )) return;
+        if (starts_with_path( name + prefix_len, len - prefix_len, no_redirect[i] )) return FALSE;
+
+    if (replace_path( attr, redir, prefix_len, system32W, syswow64W )) return TRUE;
+    if (replace_path( attr, redir, prefix_len, regeditW, syswow64_regeditW )) return TRUE;
+    return FALSE;
+}
+
+#else  /* _WIN64 */
+
+/* there are no redirects on 64-bit */
+BOOL get_redirect( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *redir )
+{
+    redir->Buffer = NULL;
+    /* CW-Bug-Id: #23185 Emulate Steam Input native hooks for native SDL */
+    return replace_steam_input_path( attr, redir );
 }
 
 #endif
@@ -3325,26 +3454,42 @@
 /* return the length of the DOS namespace prefix if any */
 static inline int get_dos_prefix_len( const UNICODE_STRING *name )
 {
+    static const WCHAR nt_prefixW[] = {'\\','?','?','\\'};
     static const WCHAR dosdev_prefixW[] = {'\\','D','o','s','D','e','v','i','c','e','s','\\'};
-
-    if (name->Length >= sizeof(nt_prefixW) &&
-        !memcmp( name->Buffer, nt_prefixW, sizeof(nt_prefixW) ))
-        return ARRAY_SIZE( nt_prefixW );
-
-    if (name->Length >= sizeof(dosdev_prefixW) &&
-        !wcsnicmp( name->Buffer, dosdev_prefixW, ARRAY_SIZE( dosdev_prefixW )))
-        return ARRAY_SIZE( dosdev_prefixW );
-
-    return 0;
+    static const WCHAR globalrootW[] = {'\\','?','?','\\','G','l','o','b','a','l','R','o','o','t'};
+    int prefix_len = 0;
+    WCHAR *prefix;
+    USHORT length;
+
+    prefix = name->Buffer;
+    length = name->Length;
+
+    if (length >= ARRAY_SIZE( globalrootW ) &&
+        !wcsnicmp( prefix, globalrootW, ARRAY_SIZE( globalrootW )))
+    {
+        WARN("Stripping off GlobalRoot prefix.\n");
+        prefix += ARRAY_SIZE( globalrootW );
+        prefix_len += ARRAY_SIZE( globalrootW );
+        length -= ARRAY_SIZE( globalrootW );
+    }
+
+    if (length >= sizeof(nt_prefixW) &&
+        !memcmp( prefix, nt_prefixW, sizeof(nt_prefixW) ))
+        prefix_len += ARRAY_SIZE( nt_prefixW );
+    else if (length >= sizeof(dosdev_prefixW) &&
+        !wcsnicmp( prefix, dosdev_prefixW, ARRAY_SIZE( dosdev_prefixW )))
+        prefix_len += ARRAY_SIZE( dosdev_prefixW );
+
+    return prefix_len;
 }
 
 
 /***********************************************************************
- *           remove_last_component
+ *           remove_last_componentA
  *
  * Remove the last component of the path. Helper for find_drive_rootA.
  */
-static unsigned int remove_last_component( const char *path, unsigned int len )
+static inline unsigned int remove_last_componentA( const char *path, unsigned int len )
 {
     int level = 0;
 
@@ -3370,29 +3515,83 @@
 }
 
 
+/***********************************************************************
+ *           find_drive_rootA
+ *
+ * Find a drive for which the root matches the beginning of the given path.
+ * This can be used to translate a Unix path into a drive + DOS path.
+ * Return value is the drive, or -1 on error. On success, ppath is modified
+ * to point to the beginning of the DOS path.
+ */
+static NTSTATUS find_drive_rootA( LPCSTR *ppath, unsigned int len, int *drive_ret )
+{
+    /* Starting with the full path, check if the device and inode match any of
+     * the wine 'drives'. If not then remove the last path component and try
+     * again. If the last component was a '..' then skip a normal component
+     * since it's a directory that's ascended back out of.
+     */
+    int drive;
+    char *buffer;
+    const char *path = *ppath;
+    struct stat st;
+    struct file_identity info[MAX_DOS_DRIVES];
+
+    /* get device and inode of all drives */
+    if (!get_drives_info( info )) return STATUS_OBJECT_PATH_NOT_FOUND;
+
+    /* strip off trailing slashes */
+    while (len > 1 && path[len - 1] == '/') len--;
+
+    /* make a copy of the path */
+    if (!(buffer = malloc( len + 1 ))) return STATUS_NO_MEMORY;
+    memcpy( buffer, path, len );
+    buffer[len] = 0;
+
+    for (;;)
+    {
+        if (!stat( buffer, &st ) && S_ISDIR( st.st_mode ))
+        {
+            /* Find the drive */
+            for (drive = 0; drive < MAX_DOS_DRIVES; drive++)
+            {
+                if ((info[drive].dev == st.st_dev) && (info[drive].ino == st.st_ino))
+                {
+                    if (len == 1) len = 0;  /* preserve root slash in returned path */
+                    TRACE( "%s -> drive %c:, root=%s, name=%s\n",
+                           debugstr_a(path), 'A' + drive, debugstr_a(buffer), debugstr_a(path + len));
+                    *ppath += len;
+                    *drive_ret = drive;
+                    free( buffer );
+                    return STATUS_SUCCESS;
+                }
+            }
+        }
+        if (len <= 1) break;  /* reached root */
+        len = remove_last_componentA( buffer, len );
+        buffer[len] = 0;
+    }
+    free( buffer );
+    return STATUS_OBJECT_PATH_NOT_FOUND;
+}
+
+
 /******************************************************************************
  *           find_file_id
  *
  * Recursively search directories from the dir queue for a given inode.
  */
-static NTSTATUS find_file_id( int root_fd, char **unix_name, ULONG *len, ULONGLONG file_id, dev_t dev, struct list *dir_queue )
+static NTSTATUS find_file_id( char **unix_name, ULONG *len, ULONGLONG file_id, dev_t dev )
 {
     unsigned int pos;
-    int dir_fd;
     DIR *dir;
     struct dirent *de;
     NTSTATUS status;
     struct stat st;
     char *name = *unix_name;
 
-    while (!(status = next_dir_in_queue( dir_queue, name )))
-    {
-        if ((dir_fd = openat( root_fd, name, O_RDONLY )) == -1) continue;
-        if (!(dir = fdopendir( dir_fd )))
-        {
-            close( dir_fd );
-            continue;
-        }
+    while (!(status = next_dir_in_queue( name )))
+    {
+        if (!(dir = opendir( name ))) continue;
         TRACE( "searching %s for %s\n", debugstr_a(name), wine_dbgstr_longlong(file_id) );
         pos = strlen( name );
         if (pos + MAX_DIR_ENTRY_LEN >= *len / sizeof(WCHAR))
@@ -3410,7 +3609,7 @@
         {
             if (!strcmp( de->d_name, "." ) || !strcmp( de->d_name, ".." )) continue;
             strcpy( name + pos, de->d_name );
-            if (fstatat( root_fd, name, &st, AT_SYMLINK_NOFOLLOW ) == -1) continue;
+            if (lstat( name, &st ) == -1) continue;
             if (st.st_dev != dev) continue;
             if (st.st_ino == file_id)
             {
@@ -3418,7 +3617,7 @@
                 return STATUS_SUCCESS;
             }
             if (!S_ISDIR( st.st_mode )) continue;
-            if ((status = add_dir_to_queue( dir_queue, name )) != STATUS_SUCCESS)
+            if ((status = add_dir_to_queue( name )) != STATUS_SUCCESS)
             {
                 closedir( dir );
                 return status;
@@ -3439,13 +3638,12 @@
                                            UNICODE_STRING *nt_name )
 {
     enum server_fd_type type;
-    int root_fd, needs_close;
+    int old_cwd, root_fd, needs_close;
     char *unix_name;
     ULONG len;
     NTSTATUS status;
     ULONGLONG file_id;
     struct stat st, root_st;
-    struct list dir_queue = LIST_INIT( dir_queue );
 
     nt_name->Buffer = NULL;
     if (attr->ObjectName->Length != sizeof(ULONGLONG)) return STATUS_OBJECT_PATH_SYNTAX_BAD;
@@ -3472,21 +3670,29 @@
         goto done;
     }
 
-    /* shortcut for ".." */
-    if (!fstatat( root_fd, "..", &st, 0 ) && st.st_dev == root_st.st_dev && st.st_ino == file_id)
-    {
-        strcpy( unix_name, ".." );
-        status = STATUS_SUCCESS;
-    }
-    else
-    {
-        status = add_dir_to_queue( &dir_queue, "." );
-        if (!status)
-            status = find_file_id( root_fd, &unix_name, &len, file_id, root_st.st_dev, &dir_queue );
-        if (!status)  /* get rid of "./" prefix */
-            memmove( unix_name, unix_name + 2, strlen(unix_name) - 1 );
-        flush_dir_queue( &dir_queue );
-    }
+    mutex_lock( &dir_mutex );
+    if ((old_cwd = open( ".", O_RDONLY )) != -1 && fchdir( root_fd ) != -1)
+    {
+        /* shortcut for ".." */
+        if (!stat( "..", &st ) && st.st_dev == root_st.st_dev && st.st_ino == file_id)
+        {
+            strcpy( unix_name, ".." );
+            status = STATUS_SUCCESS;
+        }
+        else
+        {
+            status = add_dir_to_queue( "." );
+            if (!status)
+                status = find_file_id( &unix_name, &len, file_id, root_st.st_dev );
+            if (!status)  /* get rid of "./" prefix */
+                memmove( unix_name, unix_name + 2, strlen(unix_name) - 1 );
+            flush_dir_queue();
+        }
+        if (fchdir( old_cwd ) == -1) chdir( "/" );
+    }
+    else status = errno_to_status( errno );
+    mutex_unlock( &dir_mutex );
+    if (old_cwd != -1) close( old_cwd );
 
 done:
     if (status == STATUS_SUCCESS)
@@ -3630,7 +3836,6 @@
                     return status;
                 }
             }
-            free( reparse_name );
         }
 
         /* if this is the last element, not finding it is not necessarily fatal */
@@ -3693,7 +3898,7 @@
     name     = attr->ObjectName->Buffer;
     name_len = attr->ObjectName->Length / sizeof(WCHAR);
 
-    if (!name_len || name[0] != '\\') return STATUS_OBJECT_PATH_SYNTAX_BAD;
+    if (!name || !name_len || name[0] != '\\') return STATUS_OBJECT_PATH_SYNTAX_BAD;
 
     if (!(nt_pos = get_dos_prefix_len( attr->ObjectName )))
         return STATUS_BAD_DEVICE_TYPE;  /* no DOS prefix, assume NT native name */
@@ -3783,6 +3988,135 @@
     return status;
 }
 
+
+/******************************************************************************
+ *           nt_to_unix_file_name_internal
+ *
+ * Convert a file name from NT namespace to Unix namespace.
+ *
+ * If disposition is not FILE_OPEN or FILE_OVERWRITE, the last path
+ * element doesn't have to exist; in that case STATUS_NO_SUCH_FILE is
+ * returned, but the unix name is still filled in properly.
+ */
+static NTSTATUS nt_to_unix_file_name_internal( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *nt_name,
+                                               char **name_ret, UINT disposition, BOOL open_reparse )
+{
+    enum server_fd_type type;
+    int root_fd, needs_close;
+    const WCHAR *name;
+    char *unix_name;
+    int name_len, unix_len;
+    NTSTATUS status;
+
+    if (!attr->ObjectName->Buffer && attr->ObjectName->Length)
+        return STATUS_ACCESS_VIOLATION;
+
+    if (!attr->RootDirectory)  /* without root dir fall back to normal lookup */
+        return nt_to_unix_file_name_no_root( attr, nt_name, name_ret, disposition, open_reparse, 0 );
+
+    name     = attr->ObjectName->Buffer;
+    name_len = attr->ObjectName->Length / sizeof(WCHAR);
+
+    if (name_len && name[0] == '\\') return STATUS_INVALID_PARAMETER;
+
+    unix_len = name_len * 3 + MAX_DIR_ENTRY_LEN + 3;
+    if (!(unix_name = malloc( unix_len ))) return STATUS_NO_MEMORY;
+    unix_name[0] = '.';
+
+    if (!(status = server_get_unix_fd( attr->RootDirectory, 0, &root_fd, &needs_close, &type, NULL )))
+    {
+        if (type != FD_TYPE_DIR)
+        {
+            if (needs_close) close( root_fd );
+            status = STATUS_BAD_DEVICE_TYPE;
+        }
+        else
+        {
+            status = lookup_unix_name( root_fd, attr, nt_name, 0, &unix_name, unix_len,
+                                       1, disposition, open_reparse, FALSE, 0 );
+            if (needs_close) close( root_fd );
+        }
+    }
+    else if (status == STATUS_OBJECT_TYPE_MISMATCH) status = STATUS_BAD_DEVICE_TYPE;
+
+    if (status == STATUS_SUCCESS || status == STATUS_NO_SUCH_FILE)
+    {
+        TRACE( "%s -> %s\n", debugstr_us(attr->ObjectName), debugstr_a(unix_name) );
+        *name_ret = unix_name;
+    }
+    else
+    {
+        TRACE( "%s not found in %s\n", debugstr_w(name), unix_name );
+        free( unix_name );
+    }
+    return status;
+}
+
+
+/* read the contents of an NT symlink object */
+static NTSTATUS read_nt_symlink( HANDLE root, UNICODE_STRING *name, WCHAR *target, size_t length )
+{
+    OBJECT_ATTRIBUTES attr;
+    UNICODE_STRING targetW;
+    NTSTATUS status;
+    HANDLE handle;
+
+    attr.Length = sizeof(attr);
+    attr.RootDirectory = root;
+    attr.Attributes = OBJ_CASE_INSENSITIVE;
+    attr.ObjectName = name;
+    attr.SecurityDescriptor = NULL;
+    attr.SecurityQualityOfService = NULL;
+
+    if (!(status = NtOpenSymbolicLinkObject( &handle, SYMBOLIC_LINK_QUERY, &attr )))
+    {
+        targetW.Buffer = target;
+        targetW.MaximumLength = (length - 1) * sizeof(WCHAR);
+        status = NtQuerySymbolicLinkObject( handle, &targetW, NULL );
+        NtClose( handle );
+    }
+
+    return status;
+}
+
+/* try to find dos device based on nt device name */
+static NTSTATUS nt_to_dos_device( WCHAR *name, size_t length, WCHAR *device_ret )
+{
+    static const WCHAR dosdevicesW[] = {'\\','D','o','s','D','e','v','i','c','e','s',0};
+    UNICODE_STRING dosdevW = { sizeof(dosdevicesW) - sizeof(WCHAR), sizeof(dosdevicesW), (WCHAR *)dosdevicesW };
+    WCHAR symlinkW[MAX_DIR_ENTRY_LEN];
+    OBJECT_ATTRIBUTES attr;
+    NTSTATUS status;
+    char data[1024];
+    HANDLE handle;
+    ULONG ctx = 0;
+
+    DIRECTORY_BASIC_INFORMATION *info = (DIRECTORY_BASIC_INFORMATION *)data;
+
+    attr.Length = sizeof(attr);
+    attr.RootDirectory = 0;
+    attr.ObjectName = &dosdevW;
+    attr.Attributes = OBJ_CASE_INSENSITIVE;
+    attr.SecurityDescriptor = NULL;
+    attr.SecurityQualityOfService = NULL;
+
+    status = NtOpenDirectoryObject( &handle, FILE_LIST_DIRECTORY, &attr );
+    if (status) return STATUS_BAD_DEVICE_TYPE;
+
+    while (!NtQueryDirectoryObject( handle, info, sizeof(data), TRUE, FALSE, &ctx, NULL ))
+    {
+        if (read_nt_symlink( handle, &info->ObjectName, symlinkW, MAX_DIR_ENTRY_LEN )) continue;
+        if (wcsnicmp( symlinkW, name, length )) continue;
+        if (info->ObjectName.Length != 2 * sizeof(WCHAR) || info->ObjectName.Buffer[1] != ':') continue;
+
+        *device_ret = info->ObjectName.Buffer[0];
+        NtClose( handle );
+        return STATUS_SUCCESS;
+    }
+
+    NtClose( handle );
+    return STATUS_BAD_DEVICE_TYPE;
+}
 
 /******************************************************************************
  *           nt_to_unix_file_name
@@ -3793,54 +4127,100 @@
  * element doesn't have to exist; in that case STATUS_NO_SUCH_FILE is
  * returned, but the unix name is still filled in properly.
  */
-static NTSTATUS nt_to_unix_file_name( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *nt_name,
-                                      char **name_ret, UINT disposition, BOOL open_reparse )
-{
-    enum server_fd_type type;
-    int root_fd, needs_close;
-    const WCHAR *name;
-    char *unix_name;
-    int name_len, unix_len;
+NTSTATUS nt_to_unix_file_name( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *nt_name,
+                               char **name_ret, UINT disposition, BOOL open_reparse )
+{
+    static const WCHAR systemrootW[] = {'\\','S','y','s','t','e','m','R','o','o','t','\\',0};
+    static const WCHAR dosprefixW[] = {'\\','?','?','\\'};
+    static const WCHAR deviceW[] = {'\\','D','e','v','i','c','e','\\',0};
+    WCHAR *name, *ptr, *prefix, buffer[3] = {'c',':',0};
+    UNICODE_STRING *nameW;
+    size_t offset, name_len;
     NTSTATUS status;
 
-    if (!attr->RootDirectory)  /* without root dir fall back to normal lookup */
-        return nt_to_unix_file_name_no_root( attr, nt_name, name_ret, disposition, open_reparse, 0 );
-
-    name     = attr->ObjectName->Buffer;
-    name_len = attr->ObjectName->Length / sizeof(WCHAR);
-
-    if (name_len && name[0] == '\\') return STATUS_INVALID_PARAMETER;
-
-    unix_len = name_len * 3 + MAX_DIR_ENTRY_LEN + 3;
-    if (!(unix_name = malloc( unix_len ))) return STATUS_NO_MEMORY;
-    unix_name[0] = '.';
-
-    if (!(status = server_get_unix_fd( attr->RootDirectory, 0, &root_fd, &needs_close, &type, NULL )))
-    {
-        if (type != FD_TYPE_DIR)
-        {
-            if (needs_close) close( root_fd );
-            status = STATUS_BAD_DEVICE_TYPE;
-        }
-        else
-        {
-            status = lookup_unix_name( root_fd, attr, nt_name, 0, &unix_name, unix_len,
-                                       1, disposition, open_reparse, FALSE, 0 );
-            if (needs_close) close( root_fd );
-        }
-    }
-    else if (status == STATUS_OBJECT_TYPE_MISMATCH) status = STATUS_BAD_DEVICE_TYPE;
-
-    if (status == STATUS_SUCCESS || status == STATUS_NO_SUCH_FILE)
-    {
-        TRACE( "%s -> %s\n", debugstr_us(attr->ObjectName), debugstr_a(unix_name) );
-        *name_ret = unix_name;
+    if (attr->RootDirectory) return nt_to_unix_file_name_internal( attr, nt_name, name_ret, disposition, open_reparse );
+
+    nameW = attr->ObjectName;
+
+    if (nameW->Length >= sizeof(deviceW) - sizeof(WCHAR)
+        && !wcsnicmp( nameW->Buffer, deviceW, ARRAY_SIZE(deviceW) - 1 ))
+    {
+        offset = sizeof(deviceW) / sizeof(WCHAR);
+        while (offset * sizeof(WCHAR) < nameW->Length && nameW->Buffer[ offset ] != '\\') offset++;
+        if ((status = nt_to_dos_device( nameW->Buffer, offset, buffer ))) return status;
+        prefix = buffer;
+    }
+    else if (nameW->Length >= sizeof(systemrootW) - sizeof(WCHAR) &&
+             !wcsnicmp( nameW->Buffer, systemrootW, ARRAY_SIZE(systemrootW) - 1 ))
+    {
+        offset = (sizeof(systemrootW) - 1) / sizeof(WCHAR);
+        prefix = user_shared_data->NtSystemRoot;
     }
     else
-    {
-        TRACE( "%s not found in %s\n", debugstr_w(name), unix_name );
-        free( unix_name );
-    }
+        return nt_to_unix_file_name_internal( attr, nt_name, name_ret, disposition, open_reparse );
+
+    name_len = sizeof(dosprefixW) + wcslen(prefix) * sizeof(WCHAR)
+               + sizeof(WCHAR) /* '\\' */ + nameW->Length - offset * sizeof(WCHAR) + sizeof(WCHAR);
+    if (!(name = malloc( name_len )))
+        return STATUS_NO_MEMORY;
+
+    ptr = name;
+    memcpy( ptr, dosprefixW, sizeof(dosprefixW) );
+    ptr += sizeof(dosprefixW) / sizeof(WCHAR);
+    wcscpy( ptr, prefix );
+    ptr += wcslen(ptr);
+    *ptr++ = '\\';
+    memcpy( ptr, nameW->Buffer + offset, nameW->Length - offset * sizeof(WCHAR) );
+    ptr[ nameW->Length / sizeof(WCHAR) - offset ] = 0;
+
+    nt_name->Buffer = name;
+    nt_name->Length = wcslen( name ) * sizeof(WCHAR);
+    attr->ObjectName = nt_name;
+    return nt_to_unix_file_name_internal( attr, nt_name, name_ret, disposition, open_reparse );
+}
+
+
+/******************************************************************************
+ *           wine_nt_to_unix_file_name
+ *
+ * Convert a file name from NT namespace to Unix namespace.
+ *
+ * If disposition is not FILE_OPEN or FILE_OVERWRITE, the last path
+ * element doesn't have to exist; in that case STATUS_NO_SUCH_FILE is
+ * returned, but the unix name is still filled in properly.
+ */
+NTSTATUS WINAPI wine_nt_to_unix_file_name( const OBJECT_ATTRIBUTES *attr, char *nameA, ULONG *size,
+                                          UINT disposition )
+{
+    char *buffer = NULL;
+    NTSTATUS status;
+    UNICODE_STRING nt_name;
+    OBJECT_ATTRIBUTES new_attr = *attr;
+
+    status = get_nt_and_unix_names( &new_attr, &nt_name, &buffer, disposition, FALSE );
+    if (!status || status == STATUS_NO_SUCH_FILE)
+    {
+        struct stat st1, st2;
+        char *name = buffer;
+
+        /* remove dosdevices prefix for z: drive if it points to the Unix root */
+        if (!strncmp( buffer, config_dir, strlen(config_dir) ) &&
+            !strncmp( buffer + strlen(config_dir), "/dosdevices/z:/", 15 ))
+        {
+            char *p = buffer + strlen(config_dir) + 14;
+            *p = 0;
+            if (!stat( buffer, &st1 ) && !stat( "/", &st2 ) &&
+                st1.st_dev == st2.st_dev && st1.st_ino == st2.st_ino)
+                name = p;
+            *p = '/';
+        }
+
+        if (*size > strlen(name)) strcpy( nameA, name );
+        else status = STATUS_BUFFER_TOO_SMALL;
+        *size = strlen(name) + 1;
+    }
+    free( buffer );
+    free( nt_name.Buffer );
     return status;
 }
 
@@ -3850,7 +4230,7 @@
  *
  * Get rid of . and .. components in the path.
  */
-static WCHAR *collapse_path( WCHAR *path )
+static void collapse_path( WCHAR *path )
 {
     WCHAR *p, *start, *next;
 
@@ -3920,7 +4300,6 @@
     /* remove trailing spaces and dots (yes, Windows really does that, don't ask) */
     while (p > start && (p[-1] == ' ' || p[-1] == '.')) p--;
     *p = 0;
-    return path;
 }
 
 
@@ -4064,7 +4443,7 @@
         return errno_to_status( errno );
     }
 
-    TRACE( "size %d tag %#x\n", size, data->ReparseTag );
+    TRACE( "size %d tag %#x\n", size, (int)data->ReparseTag );
 
     if (size < sizeof(*data))
     {
@@ -4162,70 +4541,12 @@
 
         /* Directory reparse tags can be opened as normal directories.
          * This is doable, but tricky, and unlikely to be needed. */
-        FIXME( "directory reparse tag %#x\n", data->ReparseTag );
+        FIXME( "directory reparse tag %#x\n", (int)data->ReparseTag );
         status = STATUS_NOT_IMPLEMENTED;
         break;
     }
 
     free( data );
-    return status;
-}
-
-
-/***********************************************************************
- *           find_drive_nt_root
- */
-static NTSTATUS find_drive_nt_root( char *unix_name, unsigned int len,
-                                    WCHAR **nt_name, UINT disposition )
-{
-    unsigned int i, pos, lenW;
-    WCHAR *buffer;
-    NTSTATUS status = STATUS_SUCCESS;
-    struct stat st;
-    struct file_identity info[MAX_DOS_DRIVES];
-
-    *nt_name = NULL;
-
-    /* get device and inode of all drives */
-    if (!get_drives_info( info )) return STATUS_OBJECT_PATH_NOT_FOUND;
-
-    /* strip off trailing slashes */
-    while (len > 1 && unix_name[len - 1] == '/') len--;
-    unix_name[len] = 0;
-
-    for (pos = len; pos; pos = remove_last_component( unix_name, pos ))
-    {
-        char prev = unix_name[pos];
-        unix_name[pos] = 0;
-        if (stat( unix_name, &st ))
-        {
-            if (pos < len) return STATUS_OBJECT_PATH_NOT_FOUND;
-            if (disposition == FILE_OPEN || disposition == FILE_OVERWRITE)
-                return STATUS_OBJECT_NAME_NOT_FOUND;
-            status = STATUS_NO_SUCH_FILE;
-            continue;
-        }
-        unix_name[pos] = prev;
-        if (!S_ISDIR( st.st_mode )) continue;
-
-        /* find the drive */
-        for (i = 0; i < MAX_DOS_DRIVES; i++)
-        {
-            if (info[i].dev != st.st_dev || info[i].ino != st.st_ino) continue;
-            while (pos < len && unix_name[pos] == '/') pos++;
-            len -= pos;
-            buffer = malloc( (len + ARRAY_SIZE(dos_prefixW) + 1) * sizeof(WCHAR) );
-            if (!buffer) return STATUS_NO_MEMORY;
-            memcpy( buffer, dos_prefixW, sizeof(dos_prefixW) );
-            buffer[4] += i;
-            lenW = ARRAY_SIZE(dos_prefixW);
-            lenW += ntdll_umbstowcs( unix_name + pos, len, buffer + lenW, len );
-            buffer[lenW] = 0;
-            *nt_name = collapse_path( buffer );
-            return status;
-        }
-        if (pos <= 1) break;
-    }
     return status;
 }
 
@@ -4233,46 +4554,57 @@
 /******************************************************************
  *           unix_to_nt_file_name
  */
-NTSTATUS unix_to_nt_file_name( const char *unix_name, WCHAR **nt, UINT disposition )
-{
+NTSTATUS unix_to_nt_file_name( const char *name, WCHAR **nt )
+{
+    static const WCHAR unix_prefixW[] = {'\\','?','?','\\','u','n','i','x',0};
+    WCHAR dos_prefixW[] = {'\\','?','?','\\','A',':','\\',0};
+    const WCHAR *prefix = unix_prefixW;
+    unsigned int lenW, lenA = strlen(name);
+    const char *path = name;
     NTSTATUS status;
     WCHAR *buffer;
-    ULONG len = strlen(unix_name) + 1;
-    char *name = strdup( unix_name );
-
-    *nt = NULL;
-    if (!name) return STATUS_NO_MEMORY;
-    status = find_drive_nt_root( name, len - 1, &buffer, disposition );
-    free( name );
-    if (status && status != STATUS_NO_SUCH_FILE) return status;
-
-    if (!buffer)  /* conversion failed, return \\?\unix path */
-    {
-        if (!(buffer = malloc( sizeof(unix_prefixW) + len * sizeof(WCHAR) ))) return STATUS_NO_MEMORY;
-        memcpy( buffer, unix_prefixW, sizeof(unix_prefixW) );
-        ntdll_umbstowcs( unix_name, len, buffer + ARRAY_SIZE(unix_prefixW), len );
-        collapse_path( buffer );
-    }
-
+    int drive;
+
+    status = find_drive_rootA( &path, lenA, &drive );
+    lenA -= path - name;
+
+    if (status == STATUS_SUCCESS)
+    {
+        while (lenA && path[0] == '/') { lenA--; path++; }
+        dos_prefixW[4] += drive;
+        prefix = dos_prefixW;
+    }
+    else if (status != STATUS_OBJECT_PATH_NOT_FOUND) return status;
+
+    lenW = wcslen( prefix );
+    if (!(buffer = malloc( (lenA + lenW + 1) * sizeof(WCHAR) ))) return STATUS_NO_MEMORY;
+    memcpy( buffer, prefix, lenW * sizeof(WCHAR) );
+    lenW += ntdll_umbstowcs( path, lenA, buffer + lenW, lenA );
+    buffer[lenW] = 0;
+    collapse_path( buffer );
     *nt = buffer;
-    return status;
+    return STATUS_SUCCESS;
 }
 
 
 /******************************************************************
- *           ntdll_get_dos_file_name
+ *           wine_unix_to_nt_file_name
  */
-NTSTATUS ntdll_get_dos_file_name( const char *unix_name, WCHAR **dos, UINT disposition )
-{
-    WCHAR *buffer;
-    NTSTATUS status = unix_to_nt_file_name( unix_name, &buffer, disposition );
-
-    if (buffer)
-    {
-        if (buffer[5] == ':') memmove( buffer, buffer + 4, (wcslen(buffer + 4) + 1) * sizeof(WCHAR) );
-        else buffer[1] = '\\';
-    }
-    *dos = buffer;
+NTSTATUS WINAPI wine_unix_to_nt_file_name( const char *name, WCHAR *buffer, ULONG *size )
+{
+    WCHAR *nt_name = NULL;
+    NTSTATUS status;
+
+    if (name[0] != '/') return STATUS_INVALID_PARAMETER;  /* relative paths are not supported */
+
+    status = unix_to_nt_file_name( name, &nt_name );
+    if (nt_name)
+    {
+        if (*size > wcslen(nt_name)) wcscpy( buffer, nt_name );
+        else status = STATUS_BUFFER_TOO_SMALL;
+        *size = wcslen(nt_name) + 1;
+        free( nt_name );
+    }
     return status;
 }
 
@@ -4317,48 +4649,22 @@
 NTSTATUS get_nt_and_unix_names( OBJECT_ATTRIBUTES *attr, UNICODE_STRING *nt_name,
                                 char **unix_name_ret, UINT disposition, BOOL open_reparse )
 {
-    ULONG lenA, lenW = attr->ObjectName->Length / sizeof(WCHAR);
     UNICODE_STRING *orig = attr->ObjectName;
     NTSTATUS status;
 
     nt_name->Buffer = NULL;
     *unix_name_ret = NULL;
 
-    if (!attr->RootDirectory && lenW > ARRAY_SIZE(unix_prefixW) &&
-        !wcsncmp( attr->ObjectName->Buffer, unix_prefixW, ARRAY_SIZE(unix_prefixW) ))
-    {
-        const WCHAR *name = attr->ObjectName->Buffer + ARRAY_SIZE(unix_prefixW);
-        char *unix_name;
-        WCHAR *buffer;
-
-        lenW -= ARRAY_SIZE(unix_prefixW);
-        *unix_name_ret = unix_name = malloc( lenW * 3 + 1 );
-        if (!unix_name) return STATUS_NO_MEMORY;
-        lenA = ntdll_wcstoumbs( name, lenW, unix_name, lenW * 3, FALSE );
-        for (ULONG i = 0; i < lenA; i++) if (unix_name[i] == '\\') unix_name[i] = '/';
-
-        status = find_drive_nt_root( unix_name, lenA, &buffer, disposition );
-        if (buffer)
-        {
-            init_unicode_string( nt_name, buffer );
-            attr->ObjectName = nt_name;
-        }
-    }
-    else
-    {
-#ifndef _WIN64
-        get_redirect( attr, nt_name );
-#endif
-        status = nt_to_unix_file_name( attr, nt_name, unix_name_ret, disposition, open_reparse );
-    }
+    get_redirect( attr, nt_name );
+    status = nt_to_unix_file_name( attr, nt_name, unix_name_ret, disposition, open_reparse );
 
     if (!status || status == STATUS_NO_SUCH_FILE)
     {
         remove_trailing_backslash( attr, nt_name );
         TRACE( "%s -> ret %x nt %s unix %s\n", debugstr_us(orig),
-               status, debugstr_us(attr->ObjectName), debugstr_a(*unix_name_ret) );
-    }
-    else TRACE( "%s -> ret %x\n", debugstr_us(orig), status );
+               (int)status, debugstr_us(attr->ObjectName), debugstr_a(*unix_name_ret) );
+    }
+    else TRACE( "%s -> ret %x\n", debugstr_us(orig), (int)status );
     return status;
 }
 
@@ -4368,138 +4674,53 @@
  *
  * Simplified version of RtlGetFullPathName_U.
  */
-NTSTATUS get_full_path( char *name, const WCHAR *curdir, UNICODE_STRING *nt_name )
-{
-    WCHAR *ret;
-    ULONG prefix_len, len = max( ARRAY_SIZE(unix_prefixW), wcslen(curdir) ) + strlen(name) + 1;
-
-    /* special case for Unix file name */
-    if (name[0] == '/' && !find_drive_nt_root( name, strlen(name), &ret, FILE_OPEN )) goto done;
-
-    if (!(ret = malloc( len * sizeof(WCHAR) ))) return STATUS_NO_MEMORY;
-
-    if (IS_SEPARATOR(name[0]) && name[1] == '?' && name[2] == '?' && IS_SEPARATOR(name[3]))  /* \??\ */
-    {
-        prefix_len = 0;
-    }
-    else if (IS_SEPARATOR(name[0]) && IS_SEPARATOR(name[1]))  /* \\ prefix */
+NTSTATUS get_full_path( const WCHAR *name, const WCHAR *curdir, WCHAR **path )
+{
+    static const WCHAR uncW[] = {'\\','?','?','\\','U','N','C','\\',0};
+    static const WCHAR devW[] = {'\\','?','?','\\',0};
+    static const WCHAR unixW[] = {'u','n','i','x'};
+    WCHAR *ret, root[] = {'\\','?','?','\\','C',':','\\',0};
+    NTSTATUS status = STATUS_SUCCESS;
+    const WCHAR *prefix;
+
+    if (IS_SEPARATOR(name[0]) && IS_SEPARATOR(name[1]))  /* \\ prefix */
     {
         if ((name[2] == '.' || name[2] == '?') && IS_SEPARATOR(name[3])) /* \\?\ device */
         {
             name += 4;
-            memcpy( ret, nt_prefixW, sizeof(nt_prefixW) );
-            prefix_len = ARRAY_SIZE(nt_prefixW);
-        }
-        else  /* UNC path */
-        {
-            name += 2;
-            memcpy( ret, unc_prefixW, sizeof(unc_prefixW) );
-            prefix_len = ARRAY_SIZE(unc_prefixW);
-        }
+            if (!wcsnicmp( name, unixW, 4 ) && IS_SEPARATOR(name[4]))  /* \\?\unix special name */
+            {
+                char *unix_name;
+                name += 4;
+                unix_name = malloc( wcslen(name) * 3 + 1 );
+                ntdll_wcstoumbs( name, wcslen(name) + 1, unix_name, wcslen(name) * 3 + 1, FALSE );
+                status = unix_to_nt_file_name( unix_name, path );
+                free( unix_name );
+                return status;
+            }
+            prefix = devW;
+        }
+        else prefix = uncW;  /* UNC path */
     }
     else if (IS_SEPARATOR(name[0]))  /* absolute path */
     {
-        memcpy( ret, dos_prefixW, sizeof(dos_prefixW) );
-        prefix_len = ARRAY_SIZE(dos_prefixW);
-        ret[4] = curdir[4];
+        root[4] = curdir[4];
+        prefix = root;
     }
     else if (name[0] && name[1] == ':')  /* drive letter */
     {
-        memcpy( ret, dos_prefixW, sizeof(dos_prefixW) );
-        prefix_len = ARRAY_SIZE(dos_prefixW);
-        ret[4] = towupper(name[0]);
+        root[4] = towupper(name[0]);
         name += 2;
-    }
-    else  /* relative path */
-    {
-        prefix_len = wcslen( curdir );
-        memcpy( ret, curdir, prefix_len * sizeof(WCHAR) );
-    }
-
-    ntdll_umbstowcs( name, strlen(name) + 1, ret + prefix_len, len - prefix_len );
+        prefix = root;
+    }
+    else prefix = curdir;  /* relative path */
+
+    ret = malloc( (wcslen(prefix) + wcslen(name) + 1) * sizeof(WCHAR) );
+    wcscpy( ret, prefix );
+    wcscat( ret, name );
     collapse_path( ret );
- done:
-    init_unicode_string( nt_name, ret );
+    *path = ret;
     return STATUS_SUCCESS;
-}
-
-
-/***********************************************************************
- *           get_nt_path
- *
- * Simplified version of RtlDosPathNameToNtPathName_U.
- */
-NTSTATUS get_nt_path( const WCHAR *name, UNICODE_STRING *nt_name )
-{
-    ULONG len = wcslen( name );
-    WCHAR *ret, *p;
-
-    nt_name->Buffer = NULL;
-    if (!(ret = p = malloc( (len + 8) * sizeof(WCHAR) ))) return STATUS_NO_MEMORY;
-
-    if (name[0] == '\\' && name[1] == '\\')
-    {
-        if ((name[2] == '.' || name[2] == '?') && name[3] == '\\')
-        {
-            memcpy( p, nt_prefixW, sizeof(nt_prefixW) );
-            p += ARRAY_SIZE( nt_prefixW );
-            name += 4;
-        }
-        else
-        {
-            memcpy( p, unc_prefixW, sizeof(unc_prefixW) );
-            p += ARRAY_SIZE( unc_prefixW );
-            name += 2;
-        }
-    }
-    else if (wcsncmp( name, nt_prefixW, ARRAY_SIZE(nt_prefixW) ))
-    {
-        memcpy( p, nt_prefixW, sizeof(nt_prefixW) );
-        p += ARRAY_SIZE( nt_prefixW );
-    }
-    wcscpy( p, name );
-    collapse_path( ret );
-    init_unicode_string( nt_name, ret );
-    return STATUS_SUCCESS;
-}
-
-
-/***********************************************************************
- *           ntdll_get_unix_file_name
- */
-NTSTATUS ntdll_get_unix_file_name( const WCHAR *dos, char **unix_name, UINT disposition )
-{
-    UNICODE_STRING nt_name, true_nt_name;
-    OBJECT_ATTRIBUTES attr;
-    char *buffer = NULL;
-    NTSTATUS status = get_nt_path( dos, &nt_name );
-
-    if (status) return status;
-    InitializeObjectAttributes( &attr, &nt_name, 0, 0, NULL );
-    status = get_nt_and_unix_names( &attr, &true_nt_name, &buffer, disposition, FALSE );
-    free( nt_name.Buffer );
-    free( true_nt_name.Buffer );
-
-    if (!status || status == STATUS_NO_SUCH_FILE)
-    {
-        /* remove dosdevices prefix for z: drive if it points to the Unix root */
-        if (!strncmp( buffer, config_dir, strlen(config_dir) ) &&
-            !strncmp( buffer + strlen(config_dir), "/dosdevices/z:/", 15 ))
-        {
-            struct stat st1, st2;
-            BOOL is_root;
-            char *p = buffer + strlen(config_dir) + 14;
-            *p = 0;
-            is_root = !stat( buffer, &st1 ) && !stat( "/", &st2 ) &&
-                      st1.st_dev == st2.st_dev && st1.st_ino == st2.st_ino;
-            *p = '/';
-            if (is_root) memmove( buffer, p, strlen(p) + 1 );
-        }
-        *unix_name = buffer;
-    }
-    else free( buffer );
-
-    return status;
 }
 
 
@@ -4599,9 +4820,9 @@
 
     TRACE( "handle=%p access=%08x name=%s objattr=%08x root=%p sec=%p io=%p alloc_size=%p "
            "attr=%08x sharing=%08x disp=%d options=%08x ea=%p.0x%08x\n",
-           handle, access, debugstr_us(attr->ObjectName), attr->Attributes,
+           handle, (int)access, debugstr_us(attr->ObjectName), (int)attr->Attributes,
            attr->RootDirectory, attr->SecurityDescriptor, io, alloc_size,
-           attributes, sharing, disposition, options, ea_buffer, ea_length );
+           (int)attributes, (int)sharing, (int)disposition, (int)options, ea_buffer, (int)ea_length );
 
     *handle = 0;
     if (!attr || !attr->ObjectName) return STATUS_INVALID_PARAMETER;
@@ -4709,7 +4930,7 @@
     struct object_attributes *objattr;
 
     TRACE( "%p %08x %p %p %08x %08x %08x %p\n",
-           handle, access, attr, io, options, quota, msg_size, timeout );
+           handle, (int)access, attr, io, (int)options, (int)quota, (int)msg_size, timeout );
 
     *handle = 0;
     if (!attr) return STATUS_INVALID_PARAMETER;
@@ -4748,9 +4969,9 @@
     if (!attr) return STATUS_INVALID_PARAMETER;
 
     TRACE( "(%p %x %s %p %x %d %x %d %d %d %d %d %d %p)\n",
-           handle, access, debugstr_us(attr->ObjectName), io, sharing, dispo,
-           options, pipe_type, read_mode, completion_mode, max_inst,
-           inbound_quota, outbound_quota, timeout );
+           handle, (int)access, debugstr_us(attr->ObjectName), io, (int)sharing, (int)dispo,
+           (int)options, (int)pipe_type, (int)read_mode, (int)completion_mode, (int)max_inst,
+           (int)inbound_quota, (int)outbound_quota, timeout );
 
     /* assume we only get relative timeout */
     if (timeout && timeout->QuadPart > 0) FIXME( "Wrong time %s\n", wine_dbgstr_longlong(timeout->QuadPart) );
@@ -4962,12 +5183,10 @@
     unsigned int options;
     unsigned int status;
 
-    TRACE( "(%p,%p,%p,0x%08x,0x%08x)\n", handle, io, ptr, len, class);
+    TRACE( "(%p,%p,%p,0x%08x,0x%08x)\n", handle, io, ptr, (int)len, class);
 
     io->Information = 0;
 
-    if (class == WineFileUnixNameInformation)
-        return server_get_file_info( handle, io, ptr, len, class );
     if (class <= 0 || class >= FileMaximumInformation)
         return io->Status = STATUS_INVALID_INFO_CLASS;
     if (!info_sizes[class])
@@ -5081,7 +5300,7 @@
         }
         break;
     case FileAttributeTagInformation:
-        if (fd_get_file_info( handle, fd, options, &st, &attr, &reparse_tag ) == -1)
+        if (fd_get_file_info( handle, fd, options, &st, &attr, (ULONG *)&reparse_tag ) == -1)
             status = errno_to_status( errno );
         else
         {
@@ -5091,7 +5310,7 @@
         }
         break;
     case FileStatInformation:
-        if (fd_get_file_info( handle, fd, options, &st, &attr, &reparse_tag ) == -1)
+        if (fd_get_file_info( handle, fd, options, &st, &attr, (ULONG *)&reparse_tag ) == -1)
             status = errno_to_status( errno );
         else if (!S_ISREG(st.st_mode) && !S_ISDIR(st.st_mode))
             status = STATUS_INVALID_INFO_CLASS;
@@ -5137,7 +5356,7 @@
     int fd, needs_close;
     unsigned int status = STATUS_SUCCESS;
 
-    TRACE( "(%p,%p,%p,0x%08x,0x%08x)\n", handle, io, ptr, len, class );
+    TRACE( "(%p,%p,%p,0x%08x,0x%08x)\n", handle, io, ptr, (int)len, class );
 
     switch (class)
     {
@@ -5160,8 +5379,8 @@
                 status = set_file_times( fd, &mtime, &atime );
 
             if (status == STATUS_SUCCESS)
-                status = fd_set_file_info( fd, info->FileAttributes,
-                                           unix_name && is_hidden_file( unix_name ));
+                status = fd_set_file_info( fd, handle, info->FileAttributes,
+                                           unix_name && is_hidden_file( unix_name ) );
 
             if (needs_close) close( fd );
             free( unix_name );
@@ -5200,6 +5419,15 @@
         }
         else status = STATUS_INVALID_PARAMETER_3;
         break;
+
+    case FileAllocationInformation:
+    {
+        const FILE_ALLOCATION_INFORMATION *info = ptr;
+
+        FIXME("FileAllocationInformation AllocationSize %p stub.\n", (void *)(ULONG_PTR)info->AllocationSize.QuadPart);
+        io->Status = STATUS_SUCCESS;
+        break;
+    }
 
     case FilePipeInformation:
         if (len >= sizeof(FILE_PIPE_INFORMATION))
@@ -5975,6 +6203,243 @@
     return status;
 }
 
+static pthread_mutex_t async_file_read_mutex = PTHREAD_MUTEX_INITIALIZER;
+static pthread_cond_t async_file_read_cond = PTHREAD_COND_INITIALIZER;
+
+struct async_file_read_job
+{
+    HANDLE handle;
+    int unix_handle;
+    int needs_close;
+    HANDLE event;
+    IO_STATUS_BLOCK *io;
+    void *buffer;
+    ULONG length;
+    LARGE_INTEGER offset;
+    DWORD thread_id;
+    LONG  cancelled;
+    struct list queue_entry;
+    struct async_file_read_job *next;
+    ULONG64 queue_time_mcs;
+};
+
+
+static struct list async_file_read_queue = LIST_INIT( async_file_read_queue );
+static struct async_file_read_job *async_file_read_running, *async_file_read_free;
+
+static void async_file_complete_io( struct async_file_read_job *job, NTSTATUS status, ULONG total )
+{
+    job->io->Status = status;
+    job->io->Information = total;
+
+    if (job->event) NtSetEvent( job->event, NULL );
+}
+
+static void *async_file_read_thread(void *dummy)
+{
+    struct async_file_read_job *job, *ptr;
+    ULONG buffer_length = 0;
+    void *buffer = NULL;
+    struct list *entry;
+    struct timespec ts;
+    NTSTATUS status;
+    ULONG64 delay;
+    ULONG total;
+    int result;
+
+    pthread_mutex_lock( &async_file_read_mutex );
+    while (1)
+    {
+        while (!(entry = list_head( &async_file_read_queue )))
+        {
+            pthread_cond_wait( &async_file_read_cond, &async_file_read_mutex );
+            continue;
+        }
+
+        job = LIST_ENTRY( entry, struct async_file_read_job, queue_entry );
+        list_remove( entry );
+
+        total = 0;
+
+        if ( job->cancelled )
+        {
+            pthread_mutex_unlock( &async_file_read_mutex );
+            status = STATUS_CANCELLED;
+            goto done;
+        }
+
+        job->next = async_file_read_running;
+        async_file_read_running = job;
+        pthread_mutex_unlock( &async_file_read_mutex );
+
+        if (!buffer_length)
+        {
+            buffer = malloc(job->length);
+            buffer_length = job->length;
+        }
+        else if (buffer_length < job->length)
+        {
+            buffer = realloc(buffer, job->length);
+            buffer_length = job->length;
+        }
+
+        while ((result = pread( job->unix_handle, buffer, job->length, job->offset.QuadPart )) == -1)
+        {
+            if (errno != EINTR)
+            {
+                status = errno_to_status( errno );
+                goto done;
+            }
+            if (job->cancelled)
+                break;
+        }
+
+        clock_gettime( CLOCK_MONOTONIC, &ts );
+        delay = ts.tv_sec * (ULONG64)1000000 + ts.tv_nsec / 1000 - job->queue_time_mcs;
+        if (delay < 1000)
+            usleep( 1000 - delay );
+        else
+            usleep( 50 );
+
+        total = result;
+        status = (total || !job->length) ? STATUS_SUCCESS : STATUS_END_OF_FILE;
+done:
+        if (job->needs_close) close( job->unix_handle );
+
+        if (!InterlockedCompareExchange(&job->cancelled, 1, 0))
+        {
+            if (status == STATUS_SUCCESS)
+                memcpy( job->buffer, buffer, total );
+
+            async_file_complete_io( job, status, total );
+        }
+
+        pthread_mutex_lock( &async_file_read_mutex );
+
+        if (status != STATUS_CANCELLED)
+        {
+            ptr = async_file_read_running;
+            if (job == ptr)
+            {
+                async_file_read_running = job->next;
+            }
+            else
+            {
+                while (ptr && ptr->next != job)
+                    ptr = ptr->next;
+
+                assert( ptr );
+                ptr->next = job->next;
+            }
+        }
+
+        job->next = async_file_read_free;
+        async_file_read_free = job;
+    }
+
+    return NULL;
+}
+
+static pthread_once_t async_file_read_once = PTHREAD_ONCE_INIT;
+
+static void async_file_read_init(void)
+{
+    pthread_t async_file_read_thread_id;
+    pthread_attr_t pthread_attr;
+
+    ERR("HACK: AC Odyssey async read workaround.\n");
+
+    pthread_attr_init( &pthread_attr );
+    pthread_attr_setscope( &pthread_attr, PTHREAD_SCOPE_SYSTEM );
+    pthread_attr_setdetachstate( &pthread_attr, PTHREAD_CREATE_DETACHED );
+
+    pthread_create( &async_file_read_thread_id, &pthread_attr, (void * (*)(void *))async_file_read_thread, NULL);
+    pthread_attr_destroy( &pthread_attr );
+}
+
+static NTSTATUS queue_async_file_read( HANDLE handle, int unix_handle, int needs_close, HANDLE event,
+                            IO_STATUS_BLOCK *io, void *buffer, ULONG length, LARGE_INTEGER *offset )
+{
+    struct async_file_read_job *job;
+    struct timespec ts;
+
+    pthread_once( &async_file_read_once, async_file_read_init );
+
+    NtResetEvent( event, NULL );
+
+    pthread_mutex_lock( &async_file_read_mutex );
+
+    if (async_file_read_free)
+    {
+        job = async_file_read_free;
+        async_file_read_free = async_file_read_free->next;
+    }
+    else
+    {
+        if (!(job = malloc( sizeof(*job) )))
+        {
+            pthread_mutex_unlock( &async_file_read_mutex );
+            return STATUS_NO_MEMORY;
+        }
+    }
+
+    job->handle = handle;
+    job->unix_handle = unix_handle;
+    job->needs_close = needs_close;
+    job->event = event;
+    job->io = io;
+    job->buffer = buffer;
+    job->length = length;
+    job->offset = *offset;
+    job->thread_id = GetCurrentThreadId();
+    job->cancelled = 0;
+    clock_gettime( CLOCK_MONOTONIC, &ts );
+    job->queue_time_mcs = ts.tv_sec * (ULONG64)1000000 + ts.tv_nsec / 1000;
+
+    list_add_tail( &async_file_read_queue, &job->queue_entry );
+
+    pthread_cond_signal( &async_file_read_cond );
+    pthread_mutex_unlock( &async_file_read_mutex );
+
+    return STATUS_PENDING;
+}
+
+static NTSTATUS cancel_async_file_read( HANDLE handle, IO_STATUS_BLOCK *io )
+{
+    DWORD thread_id = GetCurrentThreadId();
+    struct async_file_read_job *job;
+    unsigned int count = 0;
+
+    TRACE( "handle %p, io %p.\n", handle, io );
+
+    pthread_mutex_lock( &async_file_read_mutex );
+    job = async_file_read_running;
+    while (job)
+    {
+        if (((io && job->io == io)
+                || (!io && job->handle == handle && job->thread_id == thread_id))
+                && !InterlockedCompareExchange(&job->cancelled, 1, 0))
+        {
+            async_file_complete_io( job, STATUS_CANCELLED, 0 );
+            ++count;
+        }
+        job = job->next;
+    }
+
+    LIST_FOR_EACH_ENTRY( job, &async_file_read_queue, struct async_file_read_job, queue_entry )
+    {
+        if (((io && job->io == io)
+                || (!io && job->handle == handle && job->thread_id == thread_id))
+                && !InterlockedCompareExchange(&job->cancelled, 1, 0))
+        {
+            async_file_complete_io( job, STATUS_CANCELLED, 0 );
+            ++count;
+        }
+    }
+
+    pthread_mutex_unlock( &async_file_read_mutex );
+    return count ? STATUS_SUCCESS : STATUS_NOT_FOUND;
+}
 
 /******************************************************************************
  *              NtReadFile   (NTDLL.@)
@@ -5994,7 +6459,7 @@
     BOOL send_completion = FALSE, async_read, timeout_init_done = FALSE;
 
     TRACE( "(%p,%p,%p,%p,%p,%p,0x%08x,%p,%p)\n",
-           handle, event, apc, apc_user, io, buffer, length, offset, key );
+           handle, event, apc, apc_user, io, buffer, (int)length, offset, key );
 
     if (!io) return STATUS_ACCESS_VIOLATION;
 
@@ -6014,6 +6479,13 @@
         {
             status = STATUS_INVALID_PARAMETER;
             goto done;
+        }
+
+        if (ac_odyssey && async_read && length && event && !apc)
+        {
+            status = queue_async_file_read( handle, unix_handle, needs_close, event, io, buffer, length, offset );
+            needs_close = 0;
+            goto err;
         }
 
         if (offset && offset->QuadPart != FILE_USE_FILE_POINTER_POSITION)
@@ -6056,7 +6528,7 @@
            skip the synchronous read to make sure that the server starts the read
            interval timer after the first read */
         if ((status = get_io_timeouts( handle, type, length, TRUE, &timeouts ))) goto err;
-        if (timeouts.interval > 0)
+        if (timeouts.interval)
         {
             status = register_async_file_read( handle, event, apc, apc_user, iosb_ptr,
                                                buffer, total, length, FALSE );
@@ -6195,7 +6667,7 @@
     BOOL send_completion = FALSE;
 
     TRACE( "(%p,%p,%p,%p,%p,%p,0x%08x,%p,%p),partial stub!\n",
-           file, event, apc, apc_user, io, segments, length, offset, key );
+           file, event, apc, apc_user, io, segments, (int)length, offset, key );
 
     if (!io) return STATUS_ACCESS_VIOLATION;
 
@@ -6274,7 +6746,7 @@
     LARGE_INTEGER offset_eof;
 
     TRACE( "(%p,%p,%p,%p,%p,%p,0x%08x,%p,%p)\n",
-           handle, event, apc, apc_user, io, buffer, length, offset, key );
+           handle, event, apc, apc_user, io, buffer, (int)length, offset, key );
 
     if (!io) return STATUS_ACCESS_VIOLATION;
 
@@ -6494,7 +6966,7 @@
     enum server_fd_type type;
 
     TRACE( "(%p,%p,%p,%p,%p,%p,0x%08x,%p,%p),partial stub!\n",
-           file, event, apc, apc_user, io, segments, length, offset, key );
+           file, event, apc, apc_user, io, segments, (int)length, offset, key );
 
     if (length % page_size) return STATUS_INVALID_PARAMETER;
     if (!io) return STATUS_ACCESS_VIOLATION;
@@ -6570,8 +7042,8 @@
     NTSTATUS status = STATUS_NOT_SUPPORTED;
 
     TRACE( "(%p,%p,%p,%p,%p,0x%08x,%p,0x%08x,%p,0x%08x)\n",
-           handle, event, apc, apc_context, io, code,
-           in_buffer, in_size, out_buffer, out_size );
+           handle, event, apc, apc_context, io, (int)code,
+           in_buffer, (int)in_size, out_buffer, (int)out_size );
 
     /* some broken applications call this frequently with INVALID_HANDLE_VALUE,
      * and run slowly if we make a server call every time */
@@ -6658,8 +7130,8 @@
     NTSTATUS status;
 
     TRACE( "(%p,%p,%p,%p,%p,0x%08x,%p,0x%08x,%p,0x%08x)\n",
-           handle, event, apc, apc_context, io, code,
-           in_buffer, in_size, out_buffer, out_size );
+           handle, event, apc, apc_context, io, (int)code,
+           in_buffer, (int)in_size, out_buffer, (int)out_size );
 
     if (!io) return STATUS_INVALID_PARAMETER;
 
@@ -6687,7 +7159,7 @@
     case FSCTL_LOCK_VOLUME:
     case FSCTL_UNLOCK_VOLUME:
         FIXME("stub! return success - Unsupported fsctl %x (device=%x access=%x func=%x method=%x)\n",
-              code, code >> 16, (code >> 14) & 3, (code >> 2) & 0xfff, code & 3);
+              (int)code, (int)code >> 16, ((int)code >> 14) & 3, ((int)code >> 2) & 0xfff, (int)code & 3);
         status = STATUS_SUCCESS;
         break;
 
@@ -6768,10 +7240,10 @@
     enum server_fd_type type;
     int fd, needs_close;
 
-    TRACE( "(%p,0x%08x,%p,0x%08x,%p)\n", handle, flags, params, size, io );
-
-    if (flags) FIXME( "flags 0x%08x ignored\n", flags );
-    if (params || size) FIXME( "params %p/0x%08x ignored\n", params, size );
+    TRACE( "(%p,0x%08x,%p,0x%08x,%p)\n", handle, (int)flags, params, (int)size, io );
+
+    if (flags) FIXME( "flags 0x%08x ignored\n", (int)flags );
+    if (params || size) FIXME( "params %p/0x%08x ignored\n", params, (int)size );
 
     if (!io || !virtual_check_buffer_for_write( io, sizeof(*io) )) return STATUS_ACCESS_VIOLATION;
 
@@ -6821,47 +7293,31 @@
 }
 
 
-static NTSTATUS cancel_io( HANDLE handle, IO_STATUS_BLOCK *io, IO_STATUS_BLOCK *io_status,
-                           BOOL only_thread )
-{
-    HANDLE cancel_handle;
-    unsigned int status;
-
-    SERVER_START_REQ( cancel_async )
-    {
-        req->handle      = wine_server_obj_handle( handle );
-        req->iosb        = wine_server_client_ptr( io );
-        req->only_thread = only_thread;
-        if (!(status = wine_server_call( req )))
-            cancel_handle = wine_server_ptr_handle( reply->cancel_handle );
-    }
-    SERVER_END_REQ;
-
-    if (!status && cancel_handle)
-    {
-        NtWaitForSingleObject( cancel_handle, TRUE, NULL );
-        NtClose( cancel_handle );
-    }
-    else if (status == STATUS_INVALID_HANDLE)
-    {
-        return status;
-    }
-
-    io_status->Status = status;
-    io_status->Information = 0;
-
-    return status;
-}
-
-
 /**************************************************************************
  *           NtCancelIoFile   (NTDLL.@)
  */
 NTSTATUS WINAPI NtCancelIoFile( HANDLE handle, IO_STATUS_BLOCK *io_status )
 {
+    unsigned int status;
+
     TRACE( "%p %p\n", handle, io_status );
 
-    return cancel_io( handle, NULL, io_status, TRUE );
+    if (ac_odyssey && !cancel_async_file_read( handle, NULL ))
+        return (io_status->Status = STATUS_SUCCESS);
+
+    SERVER_START_REQ( cancel_async )
+    {
+        req->handle      = wine_server_obj_handle( handle );
+        req->only_thread = TRUE;
+        if (!(status = wine_server_call( req )))
+        {
+            io_status->Status = status;
+            io_status->Information = 0;
+        }
+    }
+    SERVER_END_REQ;
+
+    return status;
 }
 
 
@@ -6870,9 +7326,26 @@
  */
 NTSTATUS WINAPI NtCancelIoFileEx( HANDLE handle, IO_STATUS_BLOCK *io, IO_STATUS_BLOCK *io_status )
 {
+    unsigned int status;
+
     TRACE( "%p %p %p\n", handle, io, io_status );
 
-    return cancel_io( handle, io, io_status, FALSE );
+    if (ac_odyssey && !cancel_async_file_read( handle, io ))
+        return (io_status->Status = STATUS_SUCCESS);
+
+    SERVER_START_REQ( cancel_async )
+    {
+        req->handle = wine_server_obj_handle( handle );
+        req->iosb   = wine_server_client_ptr( io );
+        if (!(status = wine_server_call( req )))
+        {
+            io_status->Status = status;
+            io_status->Information = 0;
+        }
+    }
+    SERVER_END_REQ;
+
+    return status;
 }
 
 
@@ -7082,7 +7555,7 @@
     ULONG size = max( 4096, buffer_size );
 
     TRACE( "%p %p %p %p %p %p %u %u %d\n",
-           handle, event, apc, apc_context, iosb, buffer, buffer_size, filter, subtree );
+           handle, event, apc, apc_context, iosb, buffer, (int)buffer_size, (int)filter, subtree );
 
     if (!iosb) return STATUS_ACCESS_VIOLATION;
     if (filter == 0 || (filter & ~FILE_NOTIFY_ALL)) return STATUS_INVALID_PARAMETER;
@@ -7190,9 +7663,6 @@
                     break;
                 case D_DISK:
                     info->DeviceType = FILE_DEVICE_DISK;
-#if defined(__APPLE__)
-                    if (major(st.st_rdev) == 3 && minor(st.st_rdev) == 2) info->DeviceType = FILE_DEVICE_NULL;
-#endif
                     break;
                 case D_TTY:
                     info->DeviceType = FILE_DEVICE_SERIAL_PORT;
@@ -7203,7 +7673,7 @@
                     break;
 #endif
                 }
-                /* no special d_type for parallel ports */
+            /* no special d_type for parallel ports */
             }
         }
 #endif
@@ -7214,7 +7684,7 @@
     }
     else if (S_ISFIFO( st.st_mode ) || S_ISSOCK( st.st_mode ))
     {
-        info->DeviceType = FILE_DEVICE_NAMED_PIPE;
+        info->DeviceType = FILE_DEVICE_UNKNOWN;
     }
     else if (is_device_placeholder( fd ))
     {
@@ -7323,10 +7793,11 @@
                                               void *buffer, ULONG length,
                                               FS_INFORMATION_CLASS info_class )
 {
+    enum server_fd_type fd_type;
     int fd, needs_close;
     unsigned int status;
 
-    status = server_get_unix_fd( handle, 0, &fd, &needs_close, NULL, NULL );
+    status = server_get_unix_fd( handle, 0, &fd, &needs_close, &fd_type, NULL );
     if (status == STATUS_BAD_DEVICE_TYPE)
     {
         struct async_irp *async;
@@ -7393,7 +7864,15 @@
         {
             FILE_FS_DEVICE_INFORMATION *info = buffer;
 
-            if ((status = get_device_info( fd, info )) == STATUS_SUCCESS)
+            if (fd_type == FD_TYPE_SOCKET)
+            {
+                info->Characteristics = 0;
+                info->DeviceType = FILE_DEVICE_NAMED_PIPE;
+                status = STATUS_SUCCESS;
+            }
+            else status = get_device_info( fd, info );
+
+            if (!status)
                 io->Information = sizeof(*info);
         }
         break;
@@ -7571,7 +8050,7 @@
 NTSTATUS WINAPI NtSetVolumeInformationFile( HANDLE handle, IO_STATUS_BLOCK *io, void *info,
                                             ULONG length, FS_INFORMATION_CLASS class )
 {
-    FIXME( "(%p,%p,%p,0x%08x,0x%08x) stub\n", handle, io, info, length, class );
+    FIXME( "(%p,%p,%p,0x%08x,0x%08x) stub\n", handle, io, info, (int)length, class );
     return STATUS_SUCCESS;
 }
 
@@ -7587,7 +8066,7 @@
     NTSTATUS status;
 
     FIXME( "(%p,%p,%p,%d,%d,%p,%d,%p,%d) semi-stub\n",
-           handle, io, buffer, length, single_entry, list, list_len, index, restart );
+           handle, io, buffer, (int)length, single_entry, list, (int)list_len, index, restart );
 
     if ((status = server_get_unix_fd( handle, 0, &fd, &needs_close, NULL, NULL )))
         return status;
@@ -7605,7 +8084,7 @@
  */
 NTSTATUS WINAPI NtSetEaFile( HANDLE handle, IO_STATUS_BLOCK *io, void *buffer, ULONG length )
 {
-    FIXME( "(%p,%p,%p,%d) stub\n", handle, io, buffer, length );
+    FIXME( "(%p,%p,%p,%d) stub\n", handle, io, buffer, (int)length );
     return STATUS_ACCESS_DENIED;
 }
 
@@ -7642,7 +8121,7 @@
 {
     unsigned int status;
 
-    TRACE("(%p,0x%08x,%p,0x%08x,%p)\n", handle, info_class, ptr, len, used_len);
+    TRACE("(%p,0x%08x,%p,0x%08x,%p)\n", handle, info_class, ptr, (int)len, used_len);
 
     if (used_len) *used_len = 0;
 
@@ -7674,20 +8153,52 @@
     case ObjectNameInformation:
     {
         OBJECT_NAME_INFORMATION *p = ptr;
+        char *unix_name;
+        WCHAR *nt_name;
+
+        /* first try as a file object */
+
+        if (!(status = server_get_unix_name( handle, &unix_name )))
+        {
+            if (!(status = unix_to_nt_file_name( unix_name, &nt_name )))
+            {
+                ULONG size = (wcslen(nt_name) + 1) * sizeof(WCHAR);
+                if (len < sizeof(*p)) status = STATUS_INFO_LENGTH_MISMATCH;
+                else if (len < sizeof(*p) + size) status = STATUS_BUFFER_OVERFLOW;
+                else
+                {
+                    p->Name.Buffer = (WCHAR *)(p + 1);
+                    p->Name.Length = size - sizeof(WCHAR);
+                    p->Name.MaximumLength = size;
+                    wcscpy( p->Name.Buffer, nt_name );
+                }
+                if (used_len) *used_len = sizeof(*p) + size;
+                free( nt_name );
+            }
+            free( unix_name );
+            break;
+        }
+        else if (status != STATUS_OBJECT_TYPE_MISMATCH) break;
+
+        /* not a file, treat as a generic object */
 
         SERVER_START_REQ( get_object_name )
         {
             req->handle = wine_server_obj_handle( handle );
-            if (len > sizeof(*p) + sizeof(WCHAR))
-                wine_server_set_reply( req, p + 1, len - sizeof(*p) - sizeof(WCHAR) );
+            if (len > sizeof(*p)) wine_server_set_reply( req, p + 1, len - sizeof(*p) );
             status = wine_server_call( req );
             if (status == STATUS_SUCCESS)
             {
                 if (!reply->total)  /* no name */
                 {
-                    if (len < sizeof(*p)) status = STATUS_INFO_LENGTH_MISMATCH;
+                    if (sizeof(*p) > len) status = STATUS_INFO_LENGTH_MISMATCH;
                     else memset( p, 0, sizeof(*p) );
                     if (used_len) *used_len = sizeof(*p);
+                }
+                else if (sizeof(*p) + reply->total + sizeof(WCHAR) > len)
+                {
+                    if (used_len) *used_len = sizeof(*p) + reply->total + sizeof(WCHAR);
+                    status = STATUS_INFO_LENGTH_MISMATCH;
                 }
                 else
                 {
@@ -7699,11 +8210,6 @@
                     if (used_len) *used_len = sizeof(*p) + p->Name.MaximumLength;
                 }
             }
-            else if (status == STATUS_INFO_LENGTH_MISMATCH || status == STATUS_BUFFER_OVERFLOW)
-            {
-                if (len < sizeof(*p)) status = STATUS_INFO_LENGTH_MISMATCH;
-                if (used_len) *used_len = sizeof(*p) + reply->total + sizeof(WCHAR);
-            }
         }
         SERVER_END_REQ;
         break;
@@ -7814,7 +8320,7 @@
 {
     unsigned int status;
 
-    TRACE("(%p,0x%08x,%p,0x%08x)\n", handle, info_class, ptr, len);
+    TRACE("(%p,0x%08x,%p,0x%08x)\n", handle, info_class, ptr, (int)len);
 
     switch (info_class)
     {
