--- a/dlls/ntdll/unix/loader.c
+++ b/dlls/ntdll/unix/loader.c
@@ -90,15 +90,24 @@
 #include "winioctl.h"
 #include "winternl.h"
 #include "unix_private.h"
+#include "esync.h"
+#include "fsync.h"
 #include "wine/list.h"
 #include "ntsyscalls.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(module);
-WINE_DECLARE_DEBUG_CHANNEL(syscall);
 
-#if defined __i386__ || defined __x86_64__
-#define SO_DLLS_SUPPORTED
+#ifdef __i386__
+static const char so_dir[] = "/i386-unix";
+#elif defined(__x86_64__)
+static const char so_dir[] = "/x86_64-unix";
+#elif defined(__arm__)
+static const char so_dir[] = "/arm-unix";
+#elif defined(__aarch64__)
+static const char so_dir[] = "/aarch64-unix";
+#else
+static const char so_dir[] = "";
 #endif
 
 void *pDbgUiRemoteBreakin = NULL;
@@ -112,43 +121,25 @@
 void *p__wine_ctrl_routine = NULL;
 SYSTEM_DLL_INIT_BLOCK *pLdrSystemDllInitBlock = NULL;
 
-static void stub_syscall( const char *name )
-{
-    CONTEXT context = { .ContextFlags = CONTEXT_FULL };
-    EXCEPTION_RECORD rec =
-    {
-        .ExceptionCode = EXCEPTION_WINE_STUB,
-        .ExceptionFlags = EXCEPTION_NONCONTINUABLE,
-        .NumberParameters = 2,
-        .ExceptionInformation[0] = (ULONG_PTR)"ntdll",
-        .ExceptionInformation[1] = (ULONG_PTR)name,
-    };
-    NtGetContextThread( GetCurrentThread(), &context );
-#ifdef __i386__
-    rec.ExceptionAddress = (void *)context.Eip;
-#elif defined __x86_64__
-    rec.ExceptionAddress = (void *)context.Rip;
-#elif defined __arm__ || defined __aarch64__
-    rec.ExceptionAddress = (void *)context.Pc;
-#endif
-    NtRaiseException( &rec, &context, TRUE );
-}
-
-
-#define SYSCALL_STUB(name) static void name(void) { stub_syscall( #name ); }
-ALL_SYSCALL_STUBS
-
 static void * const syscalls[] =
 {
 #define SYSCALL_ENTRY(id,name,args) name,
-    ALL_SYSCALLS
+#ifdef _WIN64
+    ALL_SYSCALLS64
+#else
+    ALL_SYSCALLS32
+#endif
 #undef SYSCALL_ENTRY
 };
 
 static BYTE syscall_args[ARRAY_SIZE(syscalls)] =
 {
 #define SYSCALL_ENTRY(id,name,args) args,
-    ALL_SYSCALLS
+#ifdef _WIN64
+    ALL_SYSCALLS64
+#else
+    ALL_SYSCALLS32
+#endif
 #undef SYSCALL_ENTRY
 };
 
@@ -157,22 +148,6 @@
     { (ULONG_PTR *)syscalls, NULL, ARRAY_SIZE(syscalls), syscall_args }
 };
 
-static const char *ntsyscall_names[] =
-{
-#define SYSCALL_ENTRY(id,name,args) #name,
-    ALL_SYSCALLS
-#undef SYSCALL_ENTRY
-};
-
-static const char **syscall_names[4] = { ntsyscall_names };
-static const char **usercall_names;
-
-void ntdll_add_syscall_debug_info( UINT idx, const char **names, const char **user_names )
-{
-    syscall_names[idx] = names;
-    usercall_names = user_names;
-}
-
 #ifdef __GNUC__
 static void fatal_error( const char *err, ... ) __attribute__((noreturn, format(printf,1,2)));
 #endif
@@ -180,19 +155,55 @@
 static const char *bin_dir;
 static const char *dll_dir;
 static const char *ntdll_dir;
-static const char *alt_build_dir;
+static const char *wineloader;
 static SIZE_T dll_path_maxlen;
 
 const char *home_dir = NULL;
 const char *data_dir = NULL;
 const char *build_dir = NULL;
 const char *config_dir = NULL;
-const char *wineloader = NULL;
 const char **dll_paths = NULL;
 const char **system_dll_paths = NULL;
 const char *user_name = NULL;
 SECTION_IMAGE_INFORMATION main_image_info = { NULL };
 
+/* adjust an array of pointers to make them into RVAs */
+static inline void fixup_rva_ptrs( void *array, BYTE *base, unsigned int count )
+{
+    BYTE **src = array;
+    DWORD *dst = array;
+
+    for ( ; count; count--, src++, dst++) *dst = *src ? *src - base : 0;
+}
+
+/* fixup an array of RVAs by adding the specified delta */
+static inline void fixup_rva_dwords( DWORD *ptr, int delta, unsigned int count )
+{
+    for ( ; count; count--, ptr++) if (*ptr) *ptr += delta;
+}
+
+
+/* fixup an array of name/ordinal RVAs by adding the specified delta */
+static inline void fixup_rva_names( UINT_PTR *ptr, int delta )
+{
+    for ( ; *ptr; ptr++) if (!(*ptr & IMAGE_ORDINAL_FLAG)) *ptr += delta;
+}
+
+
+/* fixup RVAs in the resource directory */
+static void fixup_so_resources( IMAGE_RESOURCE_DIRECTORY *dir, BYTE *root, int delta )
+{
+    IMAGE_RESOURCE_DIRECTORY_ENTRY *entry = (IMAGE_RESOURCE_DIRECTORY_ENTRY *)(dir + 1);
+    unsigned int i;
+
+    for (i = 0; i < dir->NumberOfNamedEntries + dir->NumberOfIdEntries; i++, entry++)
+    {
+        void *ptr = root + entry->OffsetToDirectory;
+        if (entry->DataIsDirectory) fixup_so_resources( ptr, root, delta );
+        else fixup_rva_dwords( &((IMAGE_RESOURCE_DATA_ENTRY *)ptr)->OffsetToData, delta, 1 );
+    }
+}
+
 /* die on a fatal error; use only during initialization */
 static void fatal_error( const char *err, ... )
 {
@@ -345,18 +356,6 @@
 }
 
 
-static const char *get_so_dir( WORD machine )
-{
-    switch (machine)
-    {
-    case IMAGE_FILE_MACHINE_I386:  return "/i386-unix";
-    case IMAGE_FILE_MACHINE_AMD64: return "/x86_64-unix";
-    case IMAGE_FILE_MACHINE_ARMNT: return "/arm-unix";
-    case IMAGE_FILE_MACHINE_ARM64: return "/aarch64-unix";
-    default: return "";
-    }
-}
-
 static const char *get_pe_dir( WORD machine )
 {
     if (!machine) machine = current_machine;
@@ -365,9 +364,9 @@
     {
     case IMAGE_FILE_MACHINE_I386:  return "/i386-windows";
 #ifdef __aarch64__
-        case IMAGE_FILE_MACHINE_AMD64: return "/aarch64-windows";
+    case IMAGE_FILE_MACHINE_AMD64: return "/aarch64-windows";
 #else
-        case IMAGE_FILE_MACHINE_AMD64: return "/x86_64-windows";
+    case IMAGE_FILE_MACHINE_AMD64: return "/x86_64-windows";
 #endif
     case IMAGE_FILE_MACHINE_ARMNT: return "/arm-windows";
     case IMAGE_FILE_MACHINE_ARM64: return "/aarch64-windows";
@@ -375,28 +374,19 @@
     }
 }
 
-static WORD get_alt_machine( WORD machine )
-{
-    switch (machine)
-    {
-    case IMAGE_FILE_MACHINE_I386:  return IMAGE_FILE_MACHINE_AMD64;
-    case IMAGE_FILE_MACHINE_AMD64: return IMAGE_FILE_MACHINE_I386;
-    case IMAGE_FILE_MACHINE_ARMNT: return IMAGE_FILE_MACHINE_ARM64;
-    case IMAGE_FILE_MACHINE_ARM64: return IMAGE_FILE_MACHINE_ARMNT;
-    default: return machine;
-    }
-}
 
 static void set_dll_path(void)
 {
-    char *p, *path = getenv( "WINEDLLPATH" );
-    char *be_runtime = getenv( "PROTON_BATTLEYE_RUNTIME" );
-    char *eac_runtime = getenv( "PROTON_EAC_RUNTIME" );
+    char *p, *path = getenv( "WINEDLLPATH" ), *be_runtime = getenv( "PROTON_BATTLEYE_RUNTIME" ), *eac_runtime = getenv( "PROTON_EAC_RUNTIME" );
     int i, count = 0;
 
     if (path) for (p = path, count = 1; *p; p++) if (*p == ':') count++;
-    if (be_runtime) count += 2;
-    if (eac_runtime) count += 2;
+
+    if (be_runtime)
+        count += 2;
+
+    if (eac_runtime)
+        count += 2;
 
     dll_paths = malloc( (count + 2) * sizeof(*dll_paths) );
     count = 0;
@@ -415,14 +405,16 @@
         const char lib32[] = "/v1/lib/wine/";
         const char lib64[] = "/v1/lib64/wine/";
 
-        p = malloc( strlen( be_runtime ) + strlen( lib32 ) + 1 );
-        strcpy( p, be_runtime );
-        strcat( p, lib32 );
+        p = malloc( strlen(be_runtime) + strlen(lib32) + 1 );
+        strcpy(p, be_runtime);
+        strcat(p, lib32);
+
         dll_paths[count++] = p;
 
-        p = malloc( strlen( be_runtime ) + strlen( lib64 ) + 1 );
-        strcpy( p, be_runtime );
-        strcat( p, lib64 );
+        p = malloc( strlen(be_runtime) + strlen(lib64) + 1 );
+        strcpy(p, be_runtime);
+        strcat(p, lib64);
+
         dll_paths[count++] = p;
     }
 
@@ -431,14 +423,16 @@
         const char lib32[] = "/v2/lib32/";
         const char lib64[] = "/v2/lib64/";
 
-        p = malloc( strlen( eac_runtime ) + strlen( lib32 ) + 1 );
-        strcpy( p, eac_runtime );
-        strcat( p, lib32 );
+        p = malloc( strlen(eac_runtime) + strlen(lib32) + 1 );
+        strcpy(p, eac_runtime);
+        strcat(p, lib32);
+
         dll_paths[count++] = p;
 
-        p = malloc( strlen( eac_runtime ) + strlen( lib64 ) + 1 );
-        strcpy( p, eac_runtime );
-        strcat( p, lib64 );
+        p = malloc( strlen(eac_runtime) + strlen(lib64) + 1 );
+        strcpy(p, eac_runtime);
+        strcat(p, lib64);
+
         dll_paths[count++] = p;
     }
 
@@ -514,37 +508,38 @@
 static void init_paths( char *argv[] )
 {
     Dl_info info;
-    const char *loader_name = "wine";
-    const char *slash;
-    char *loader_dir, *wow64_loader, *env;
+    char *basename, *env;
 
-    if (argv && argv[0] && argv[0][0])
-    {
-        slash = strrchr( argv[0], '/' );
-        loader_name = slash ? slash + 1 : argv[0];
-    }
+    if ((basename = strrchr( argv[0], '/' ))) basename++;
+    else basename = argv[0];
 
     if (!dladdr( init_paths, &info ) || !(ntdll_dir = realpath_dirname( info.dli_fname )))
         fatal_error( "cannot get path to ntdll.so\n" );
 
-    if ((build_dir = remove_tail( ntdll_dir, "/dlls/ntdll" )))
+    if (!(build_dir = remove_tail( ntdll_dir, "/dlls/ntdll" )))
     {
-        loader_dir = build_path( build_dir, "loader" );
-        wineloader = build_path( loader_dir, loader_name );
-        free( loader_dir );
-        wow64_loader = build_path( build_dir, "loader-wow64" );
-        alt_build_dir = realpath_dirname( wow64_loader );
-        free( wow64_loader );
-    }
-    else
-    {
-        if (!(dll_dir = remove_tail( ntdll_dir, get_so_dir(current_machine) ))) dll_dir = ntdll_dir;
-#if defined(__linux__) && !defined(__ANDROID__)
+        if (!(dll_dir = remove_tail( ntdll_dir, so_dir ))) dll_dir = ntdll_dir;
+#if (defined(__linux__) && !defined(__ANDROID__)) || defined(__FreeBSD_kernel__) || defined(__NetBSD__)
         bin_dir = realpath_dirname( "/proc/self/exe" );
+#elif defined (__FreeBSD__) || defined(__DragonFly__)
+        {
+            static int pathname[] = { CTL_KERN, KERN_PROC, KERN_PROC_PATHNAME, -1 };
+            size_t path_size = PATH_MAX;
+            char *path = malloc( path_size );
+            if (path && !sysctl( pathname, ARRAY_SIZE(pathname), path, &path_size, NULL, 0 ))
+                bin_dir = realpath_dirname( path );
+            free( path );
+        }
 #endif
         if (!bin_dir) bin_dir = build_relative_path( dll_dir, LIBDIR "/wine", BINDIR );
-        data_dir = build_relative_path( dll_dir, LIBDIR "/wine", DATADIR "/wine" );
-        wineloader = build_path( ntdll_dir, loader_name );
+        data_dir = build_relative_path( bin_dir, BINDIR, DATADIR "/wine" );
+        wineloader = build_path( bin_dir, basename );
+    }
+    else
+    {
+        char *dirname = build_path( build_dir, "loader" );
+        wineloader = build_path( dirname, basename );
+        free(dirname);
     }
 
     asprintf( &env, "WINELOADER=%s", wineloader );
@@ -562,27 +557,25 @@
  */
 char *get_alternate_wineloader( WORD machine )
 {
-    const char *arch;
-    BOOL force_wow64 = (arch = getenv( "WINEARCH" )) && !strcmp( arch, "wow64" );
     char *ret = NULL;
 
-    if (is_win64)
-    {
-        if (force_wow64) return NULL;
-        if (machine != get_alt_machine( current_machine )) return NULL;
-    }
-    else
+    if (machine == current_machine) return NULL;
+
+    /* try the 64-bit loader */
+    if (current_machine == IMAGE_FILE_MACHINE_I386 && machine == IMAGE_FILE_MACHINE_AMD64)
     {
-        if (!force_wow64 && machine == current_machine) return NULL;
-        machine = get_alt_machine( current_machine );
-    }
+        size_t len = strlen(wineloader);
 
-    if (!build_dir)
-        asprintf( &ret, "%s%s/wine", dll_dir, get_so_dir( machine ));
-    else if (alt_build_dir)
-        asprintf( &ret, "%s/loader/wine", alt_build_dir );
+        if (len <= 2 || strcmp( wineloader + len - 2, "64" ))
+        {
+            ret = malloc( len + 3 );
+            strcpy( ret, wineloader );
+            strcat( ret, "64" );
+        }
+        return ret;
+    }
 
-    return ret;
+    return remove_tail( wineloader, "64" );
 }
 
 
@@ -599,6 +592,7 @@
     argv[0] = malloc( p - argv[1] + strlen(preloader) + 1 );
     memcpy( argv[0], argv[1], p - argv[1] );
     strcpy( argv[0] + (p - argv[1]), preloader );
+
 #ifdef __APPLE__
     {
         posix_spawnattr_t attr;
@@ -617,10 +611,6 @@
 /* exec the appropriate wine loader for the specified machine */
 static NTSTATUS loader_exec( char **argv, WORD machine )
 {
-    static char noexec[] = "WINELOADERNOEXEC=1";
-
-    putenv( noexec );
-
     if (((argv[1] = get_alternate_wineloader( machine )))) preloader_exec( argv );
 
     argv[1] = strdup( wineloader );
@@ -766,119 +756,6 @@
     return TRUE;
 }
 
-void trace_syscall( UINT id, ULONG_PTR *args, ULONG len )
-{
-    UINT idx = (id >> 12) & 3, num = id & 0xfff;
-    const char **names = syscall_names[idx];
-
-    if (names && names[num])
-        TRACE_(syscall)( "\1SysCall  %s(", names[num] );
-    else
-        TRACE_(syscall)( "\1SysCall  %04x(", id );
-
-    len /= sizeof(ULONG_PTR);
-    for (ULONG i = 0; i < len; i++)
-    {
-        TRACE_(syscall)( "%08lx", args[i] );
-        if (i < len - 1) TRACE_(syscall)( "," );
-    }
-    TRACE_(syscall)( ")\n" );
-}
-
-void trace_sysret( UINT id, ULONG_PTR retval )
-{
-    UINT idx = (id >> 12) & 3, num = id & 0xfff;
-    const char **names = syscall_names[idx];
-
-    if (names && names[num])
-        TRACE_(syscall)( "\1SysRet   %s() retval=%08lx\n", names[num], retval );
-    else
-        TRACE_(syscall)( "\1SysRet   %04x() retval=%08lx\n", id, retval );
-}
-
-void trace_usercall( UINT id, ULONG_PTR *args, ULONG len )
-{
-    if (usercall_names)
-        TRACE_(syscall)("\1UserCall %s(%p,%u)\n", usercall_names[id], args, len );
-    else
-        TRACE_(syscall)("\1UserCall %04x(%p,%u)\n", id, args, len );
-}
-
-void trace_userret( void *ret_ptr, ULONG len, NTSTATUS status, UINT id )
-{
-    if (usercall_names)
-        TRACE_(syscall)("\1UserRet  %s(%p,%u) retval=%08x\n", usercall_names[id], ret_ptr, len, status );
-    else
-        TRACE_(syscall)("\1UserRet  %04x(%p,%u) retval=%08x\n", id, ret_ptr, len, status );
-}
-
-#ifdef SO_DLLS_SUPPORTED
-
-/* adjust an array of pointers to make them into RVAs */
-static inline void fixup_rva_ptrs( void *array, BYTE *base, unsigned int count )
-{
-    BYTE **src = array;
-    DWORD *dst = array;
-
-    for ( ; count; count--, src++, dst++) *dst = *src ? *src - base : 0;
-}
-
-/* fixup an array of RVAs by adding the specified delta */
-static inline void fixup_rva_dwords( DWORD *ptr, int delta, unsigned int count )
-{
-    for ( ; count; count--, ptr++) if (*ptr) *ptr += delta;
-}
-
-
-/* fixup an array of name/ordinal RVAs by adding the specified delta */
-static inline void fixup_rva_names( UINT_PTR *ptr, int delta )
-{
-    for ( ; *ptr; ptr++) if (!(*ptr & IMAGE_ORDINAL_FLAG)) *ptr += delta;
-}
-
-
-/* fixup RVAs in the resource directory */
-static void fixup_so_resources( IMAGE_RESOURCE_DIRECTORY *dir, BYTE *root, int delta )
-{
-    IMAGE_RESOURCE_DIRECTORY_ENTRY *entry = (IMAGE_RESOURCE_DIRECTORY_ENTRY *)(dir + 1);
-    unsigned int i;
-
-    for (i = 0; i < dir->NumberOfNamedEntries + dir->NumberOfIdEntries; i++, entry++)
-    {
-        void *ptr = root + entry->OffsetToDirectory;
-        if (entry->DataIsDirectory) fixup_so_resources( ptr, root, delta );
-        else fixup_rva_dwords( &((IMAGE_RESOURCE_DATA_ENTRY *)ptr)->OffsetToData, delta, 1 );
-    }
-}
-
-/***********************************************************************
- *           fill_builtin_image_info
- */
-static void fill_builtin_image_info( void *module, struct pe_image_info *info )
-{
-    const IMAGE_DOS_HEADER *dos = (const IMAGE_DOS_HEADER *)module;
-    const IMAGE_NT_HEADERS *nt = (IMAGE_NT_HEADERS *)((const BYTE *)dos + dos->e_lfanew);
-
-    memset( info, 0, sizeof(*info) );
-    info->base            = nt->OptionalHeader.ImageBase;
-    info->entry_point     = nt->OptionalHeader.AddressOfEntryPoint;
-    info->map_size        = nt->OptionalHeader.SizeOfImage;
-    info->stack_size      = nt->OptionalHeader.SizeOfStackReserve;
-    info->stack_commit    = nt->OptionalHeader.SizeOfStackCommit;
-    info->subsystem       = nt->OptionalHeader.Subsystem;
-    info->subsystem_minor = nt->OptionalHeader.MinorSubsystemVersion;
-    info->subsystem_major = nt->OptionalHeader.MajorSubsystemVersion;
-    info->osversion_major = nt->OptionalHeader.MajorOperatingSystemVersion;
-    info->osversion_minor = nt->OptionalHeader.MinorOperatingSystemVersion;
-    info->image_charact   = nt->FileHeader.Characteristics;
-    info->dll_charact     = nt->OptionalHeader.DllCharacteristics;
-    info->machine         = nt->FileHeader.Machine;
-    info->contains_code   = TRUE;
-    info->wine_builtin    = TRUE;
-    info->header_size     = nt->OptionalHeader.SizeOfHeaders;
-    info->file_size       = nt->OptionalHeader.SizeOfImage;
-    info->checksum        = nt->OptionalHeader.CheckSum;
-}
 
 /*************************************************************************
  *		map_so_dll
@@ -893,18 +770,15 @@
     IMAGE_NT_HEADERS *nt;
     IMAGE_SECTION_HEADER *sec;
     BYTE *addr = (BYTE *)module;
-    DWORD code_start, code_end, data_start, data_end;
-    DWORD align_mask = nt_descr->OptionalHeader.SectionAlignment - 1;
+    DWORD code_start, code_end, data_start, data_end, align_mask;
     int delta, nb_sections = 2;  /* code + data */
     unsigned int i;
-
-    code_start = (sizeof(IMAGE_DOS_HEADER)
+    DWORD size = (sizeof(IMAGE_DOS_HEADER)
                   + sizeof(builtin_signature)
                   + sizeof(IMAGE_NT_HEADERS)
-                  + nb_sections * sizeof(IMAGE_SECTION_HEADER)
-                  + align_mask) & ~align_mask;
+                  + nb_sections * sizeof(IMAGE_SECTION_HEADER));
 
-    if (anon_mmap_fixed( addr, code_start, PROT_READ | PROT_WRITE, 0 ) != addr) return STATUS_NO_MEMORY;
+    if (anon_mmap_fixed( addr, size, PROT_READ | PROT_WRITE, 0 ) != addr) return STATUS_NO_MEMORY;
 
     dos = (IMAGE_DOS_HEADER *)addr;
     nt  = (IMAGE_NT_HEADERS *)((BYTE *)(dos + 1) + sizeof(builtin_signature));
@@ -926,6 +800,8 @@
     *nt = *nt_descr;
 
     delta      = (const BYTE *)nt_descr - addr;
+    align_mask = nt->OptionalHeader.SectionAlignment - 1;
+    code_start = (size + align_mask) & ~align_mask;
     data_start = delta & ~align_mask;
 #ifdef __APPLE__
     {
@@ -1045,6 +921,84 @@
     return STATUS_SUCCESS;
 }
 
+static ULONG_PTR find_ordinal_export( HMODULE module, const IMAGE_EXPORT_DIRECTORY *exports, DWORD ordinal )
+{
+    const DWORD *functions = (const DWORD *)((BYTE *)module + exports->AddressOfFunctions);
+
+    if (ordinal >= exports->NumberOfFunctions) return 0;
+    if (!functions[ordinal]) return 0;
+    return (ULONG_PTR)module + functions[ordinal];
+}
+
+static ULONG_PTR find_named_export( HMODULE module, const IMAGE_EXPORT_DIRECTORY *exports,
+                                    const char *name )
+{
+    const WORD *ordinals = (const WORD *)((BYTE *)module + exports->AddressOfNameOrdinals);
+    const DWORD *names = (const DWORD *)((BYTE *)module + exports->AddressOfNames);
+    int min = 0, max = exports->NumberOfNames - 1;
+
+    while (min <= max)
+    {
+        int res, pos = (min + max) / 2;
+        char *ename = (char *)module + names[pos];
+        if (!(res = strcmp( ename, name ))) return find_ordinal_export( module, exports, ordinals[pos] );
+        if (res > 0) max = pos - 1;
+        else min = pos + 1;
+    }
+    return 0;
+}
+
+static inline void *get_rva( void *module, ULONG_PTR addr )
+{
+    return (BYTE *)module + addr;
+}
+
+static const void *get_module_data_dir( HMODULE module, ULONG dir, ULONG *size )
+{
+    const IMAGE_NT_HEADERS *nt = get_rva( module, ((IMAGE_DOS_HEADER *)module)->e_lfanew );
+    const IMAGE_DATA_DIRECTORY *data;
+
+    if (nt->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR64_MAGIC)
+        data = &((const IMAGE_NT_HEADERS64 *)nt)->OptionalHeader.DataDirectory[dir];
+    else if (nt->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR32_MAGIC)
+        data = &((const IMAGE_NT_HEADERS32 *)nt)->OptionalHeader.DataDirectory[dir];
+    else
+        return NULL;
+    if (!data->VirtualAddress || !data->Size) return NULL;
+    if (size) *size = data->Size;
+    return get_rva( module, data->VirtualAddress );
+}
+
+/***********************************************************************
+ *           fill_builtin_image_info
+ */
+static void fill_builtin_image_info( void *module, struct pe_image_info *info )
+{
+    const IMAGE_DOS_HEADER *dos = (const IMAGE_DOS_HEADER *)module;
+    const IMAGE_NT_HEADERS *nt = (IMAGE_NT_HEADERS *)((const BYTE *)dos + dos->e_lfanew);
+
+    memset( info, 0, sizeof(*info) );
+    info->base            = nt->OptionalHeader.ImageBase;
+    info->entry_point     = nt->OptionalHeader.AddressOfEntryPoint;
+    info->map_size        = nt->OptionalHeader.SizeOfImage;
+    info->stack_size      = nt->OptionalHeader.SizeOfStackReserve;
+    info->stack_commit    = nt->OptionalHeader.SizeOfStackCommit;
+    info->subsystem       = nt->OptionalHeader.Subsystem;
+    info->subsystem_minor = nt->OptionalHeader.MinorSubsystemVersion;
+    info->subsystem_major = nt->OptionalHeader.MajorSubsystemVersion;
+    info->osversion_major = nt->OptionalHeader.MajorOperatingSystemVersion;
+    info->osversion_minor = nt->OptionalHeader.MinorOperatingSystemVersion;
+    info->image_charact   = nt->FileHeader.Characteristics;
+    info->dll_charact     = nt->OptionalHeader.DllCharacteristics;
+    info->machine         = nt->FileHeader.Machine;
+    info->contains_code   = TRUE;
+    info->wine_builtin    = TRUE;
+    info->header_size     = nt->OptionalHeader.SizeOfHeaders;
+    info->file_size       = nt->OptionalHeader.SizeOfImage;
+    info->checksum        = nt->OptionalHeader.CheckSum;
+}
+
+
 /***********************************************************************
  *           dlopen_dll
  */
@@ -1132,135 +1086,214 @@
     return status;
 }
 
-/* check if the library is the correct architecture */
-/* only returns false for a valid library of the wrong arch */
-static int check_library_arch( int fd )
+
+/**********************************************************************
+ *      __wine_get_unix_env
+ */
+NTSTATUS WINAPI wine_get_unix_env( void *args )
 {
-#ifdef __APPLE__
-    struct  /* Mach-O header */
-    {
-        unsigned int magic;
-        unsigned int cputype;
-    } header;
+    struct wine_get_unix_env_params *params = args;
+    unsigned int len;
+    char *s;
+
+    if (!(s = getenv( params->name ))) return STATUS_VARIABLE_NOT_FOUND;
+    len = strlen( s ) + 1;
+    if (len > params->buffer_len) return STATUS_BUFFER_TOO_SMALL;
+    memcpy( params->val, s, len );
+    return STATUS_SUCCESS;
+}
 
-    if (read( fd, &header, sizeof(header) ) != sizeof(header)) return 1;
-    if (header.magic != 0xfeedface) return 1;
-    if (sizeof(void *) == sizeof(int)) return !(header.cputype >> 24);
-    else return (header.cputype >> 24) == 1; /* CPU_ARCH_ABI64 */
-#else
-    struct  /* ELF header */
-    {
-        unsigned char magic[4];
-        unsigned char class;
-        unsigned char data;
-        unsigned char version;
-    } header;
 
-    if (read( fd, &header, sizeof(header) ) != sizeof(header)) return 1;
-    if (memcmp( header.magic, "\177ELF", 4 )) return 1;
-    if (header.version != 1 /* EV_CURRENT */) return 1;
-#ifdef WORDS_BIGENDIAN
-    if (header.data != 2 /* ELFDATA2MSB */) return 1;
-#else
-    if (header.data != 1 /* ELFDATA2LSB */) return 1;
-#endif
-    if (sizeof(void *) == sizeof(int)) return header.class == 1; /* ELFCLASS32 */
-    else return header.class == 2; /* ELFCLASS64 */
-#endif
+/**********************************************************************
+ *      __wine_set_unix_env
+ */
+NTSTATUS WINAPI wine_set_unix_env( void *args )
+{
+    struct wine_set_unix_env_params *params = args;
+
+    if (!params->val) unsetenv( params->name );
+    else setenv( params->name, params->val, 1 );
+    return 0;
 }
 
-/***********************************************************************
- *           open_builtin_so_file
+
+/**********************************************************************
+ *      __wine_dbg_ftrace
  */
-static NTSTATUS open_builtin_so_file( char *name, OBJECT_ATTRIBUTES *attr, void **module,
-                                      SECTION_IMAGE_INFORMATION *image_info, USHORT search_machine,
-                                      USHORT load_machine, BOOL prefer_native )
+static NTSTATUS unix__wine_dbg_ftrace( void *args )
 {
-    NTSTATUS status = STATUS_DLL_NOT_FOUND;
-    int fd;
-    char *end = name + strlen( name );
+    struct wine_dbg_ftrace_params *params = args;
 
-    if (search_machine != current_machine) return status;
-    if (load_machine && load_machine != current_machine) return status;
+    return __wine_dbg_ftrace( params->str, params->len, params->ctx );
+}
 
-    *module = NULL;
-    strcpy( end, ".so" );
-    if ((fd = open( name, O_RDONLY )) == -1) goto done;
 
-    if (check_library_arch( fd ))
-    {
-        struct pe_image_info info;
+static void *steamclient_srcs[128];
+static void *steamclient_tgts[128];
+static int steamclient_count;
 
-        status = dlopen_dll( name, attr->ObjectName, module, &info, prefer_native );
-        if (!status) virtual_fill_image_information( &info, image_info );
-        else if (status != STATUS_IMAGE_ALREADY_LOADED)
-        {
-            ERR( "failed to load .so lib %s\n", debugstr_a(name) );
-            status = STATUS_PROCEDURE_NOT_FOUND;
-        }
+void *steamclient_handle_fault( LPCVOID addr, DWORD err )
+{
+    int i;
+
+    if (!(err & EXCEPTION_EXECUTE_FAULT)) return NULL;
+
+    for (i = 0; i < steamclient_count; ++i)
+    {
+        if (addr == steamclient_srcs[i])
+            return steamclient_tgts[i];
     }
-    else status = STATUS_NOT_SUPPORTED;
 
-    close( fd );
- done:
-    *end = 0;
-    return status;
+    return NULL;
 }
 
-/***********************************************************************
- *           open_main_image_so_file
- */
-static NTSTATUS open_main_image_so_file( const char *name, UNICODE_STRING *nt_name, void **module,
-                                         SECTION_IMAGE_INFORMATION *image_info )
+static void steamclient_write_jump_x64(void *src_addr, ULONGLONG tgt_addr)
 {
-    struct pe_image_info pe_info;
-    NTSTATUS status;
+    static const char mov[] = {0x48, 0xb8};
+    static const char jmp[] = {0xff, 0xe0};
+    memcpy(src_addr, mov, sizeof(mov));
+    memcpy((char *)src_addr + sizeof(mov), &tgt_addr, sizeof(tgt_addr));
+    memcpy((char *)src_addr + sizeof(mov) + sizeof(tgt_addr), jmp, sizeof(jmp));
+}
 
-    /* remove .so extension from Windows name */
-    if (nt_name->Length > 3 * sizeof(WCHAR))
-    {
-        static const WCHAR soW[] = {'.','s','o',0};
-        WCHAR *p = nt_name->Buffer + nt_name->Length / sizeof(WCHAR);
-        if (!wcsicmp( p - 3, soW ))
+static void steamclient_write_jump_x86(void *src_addr, ULONG tgt_addr)
+{
+    static const char mov[] = {0xb8};
+    static const char jmp[] = {0xff, 0xe0};
+    memcpy(src_addr, mov, sizeof(mov));
+    memcpy((char *)src_addr + sizeof(mov), &tgt_addr, sizeof(tgt_addr));
+    memcpy((char *)src_addr + sizeof(mov) + sizeof(tgt_addr), jmp, sizeof(jmp));
+}
+
+static NTSTATUS steamclient_setup_trampolines( void *args )
+{
+    static int noexec_cached = -1;
+
+    struct steamclient_setup_trampolines_params *params = args;
+    HMODULE src_mod = params->src_mod, tgt_mod = params->tgt_mod;
+    SYSTEM_BASIC_INFORMATION info;
+    IMAGE_NT_HEADERS *src_nt = get_rva( src_mod, ((IMAGE_DOS_HEADER *)src_mod)->e_lfanew );
+    IMAGE_NT_HEADERS *tgt_nt = get_rva( tgt_mod, ((IMAGE_DOS_HEADER *)tgt_mod)->e_lfanew );
+    IMAGE_SECTION_HEADER *src_sec = IMAGE_FIRST_SECTION( src_nt );
+    BOOL x64 = src_nt->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR64_MAGIC;
+    const IMAGE_EXPORT_DIRECTORY *src_exp, *tgt_exp;
+    const DWORD *names;
+    SIZE_T size;
+    void *addr, *src_addr, *tgt_addr;
+    char *name, *wsne;
+    UINT_PTR page_mask;
+    int i;
+
+    if (noexec_cached == -1)
+        noexec_cached = (wsne = getenv("WINESTEAMNOEXEC")) && atoi(wsne);
+
+    virtual_get_system_info( &info, !!NtCurrentTeb()->WowTebOffset );
+    page_mask = info.PageSize - 1;
+
+    for (i = 0; i < src_nt->FileHeader.NumberOfSections; ++i)
+    {
+        if (memcmp(src_sec[i].Name, ".text", 5)) continue;
+        addr = (void *)(((UINT_PTR)src_mod + src_sec[i].VirtualAddress) & ~page_mask);
+        size = (src_sec[i].Misc.VirtualSize + page_mask) & ~page_mask;
+        if (noexec_cached) mprotect(addr, size, PROT_READ);
+        else mprotect(addr, size, PROT_READ|PROT_WRITE|PROT_EXEC);
+    }
+
+    src_exp = get_module_data_dir( src_mod, IMAGE_FILE_EXPORT_DIRECTORY, NULL );
+    tgt_exp = get_module_data_dir( tgt_mod, IMAGE_FILE_EXPORT_DIRECTORY, NULL );
+    names = (const DWORD *)((UINT_PTR)src_mod + src_exp->AddressOfNames);
+    for (i = 0; i < src_exp->NumberOfNames; ++i)
+    {
+        if (!names[i] || !(name = (char *)((UINT_PTR)src_mod + names[i]))) continue;
+        if (!(src_addr = (void *)find_named_export(src_mod, src_exp, name))) continue;
+        if (!(tgt_addr = (void *)find_named_export(tgt_mod, tgt_exp, name))) continue;
+        assert(steamclient_count < ARRAY_SIZE(steamclient_srcs));
+        steamclient_srcs[steamclient_count] = src_addr;
+        steamclient_tgts[steamclient_count] = tgt_addr;
+        if (!noexec_cached)
         {
-            p[-3] = 0;
-            nt_name->Length -= 3 * sizeof(WCHAR);
+            if (x64) steamclient_write_jump_x64( src_addr, (ULONG_PTR)tgt_addr );
+            else steamclient_write_jump_x86( src_addr, PtrToUlong(tgt_addr) );
         }
+        else steamclient_count++;
     }
-    status = dlopen_dll( name, nt_name, module, &pe_info, FALSE );
-    if (!status) virtual_fill_image_information( &pe_info, image_info );
-    return status;
-}
 
-extern NTSTATUS unwind_builtin_dll( void *args );
+    if (x64)
+    {
+        IMAGE_NT_HEADERS64 *src_nt64 = (IMAGE_NT_HEADERS64 *)src_nt, *tgt_nt64 = (IMAGE_NT_HEADERS64 *)tgt_nt;
+        src_addr = (void *)((UINT_PTR)src_mod + src_nt64->OptionalHeader.AddressOfEntryPoint);
+        tgt_addr = (void *)((UINT_PTR)tgt_mod + tgt_nt64->OptionalHeader.AddressOfEntryPoint);
+    }
+    else
+    {
+        IMAGE_NT_HEADERS32 *src_nt32 = (IMAGE_NT_HEADERS32 *)src_nt, *tgt_nt32 = (IMAGE_NT_HEADERS32 *)tgt_nt;
+        src_addr = (void *)((UINT_PTR)src_mod + src_nt32->OptionalHeader.AddressOfEntryPoint);
+        tgt_addr = (void *)((UINT_PTR)tgt_mod + tgt_nt32->OptionalHeader.AddressOfEntryPoint);
+    }
 
-#else /* SO_DLLS_SUPPORTED */
+    assert(steamclient_count < ARRAY_SIZE(steamclient_srcs));
+    steamclient_srcs[steamclient_count] = src_addr;
+    steamclient_tgts[steamclient_count] = tgt_addr;
+    if (!noexec_cached)
+    {
+        if (x64) steamclient_write_jump_x64( src_addr, (ULONG_PTR)tgt_addr );
+        else steamclient_write_jump_x86( src_addr, PtrToUlong(tgt_addr) );
+    }
+    else steamclient_count++;
 
-static NTSTATUS open_builtin_so_file( char *name, OBJECT_ATTRIBUTES *attr, void **module,
-                                      SECTION_IMAGE_INFORMATION *image_info, USHORT search_machine,
-                                      USHORT load_machine, BOOL prefer_native )
-{
-    return STATUS_DLL_NOT_FOUND;
+    return STATUS_SUCCESS;
 }
 
-static NTSTATUS open_main_image_so_file( const char *name, UNICODE_STRING *nt_name, void **module,
-                                         SECTION_IMAGE_INFORMATION *image_info )
+static BOOL debugstr_pc_impl( void *pc, char *buffer, unsigned int size )
 {
-    return STATUS_INVALID_IMAGE_FORMAT;
+    unsigned int len;
+    char *s = buffer;
+    Dl_info info;
+
+    snprintf( s, size, "%p:", pc );
+    if (!dladdr( pc, &info )) return FALSE;
+
+    s += (len = strlen( s ));
+    size -= len;
+    snprintf( s, size, " %s + %#zx", info.dli_fname, (char *)pc - (char *)info.dli_fbase );
+    if (info.dli_sname)
+    {
+        s += (len = strlen( s ));
+        size -= len;
+        snprintf( s, size, " (%s + %#zx)", info.dli_sname, (char *)pc - (char *)info.dli_saddr );
+    }
+    return TRUE;
 }
 
-static NTSTATUS load_so_dll( void *args )
+static NTSTATUS debugstr_pc( void *args )
 {
-    return STATUS_INVALID_IMAGE_FORMAT;
+    struct debugstr_pc_args *params = args;
+
+    return debugstr_pc_impl( params->pc, params->buffer, params->size ) ? STATUS_SUCCESS : STATUS_UNSUCCESSFUL;
 }
 
-static NTSTATUS unwind_builtin_dll( void *args )
+const char * wine_debuginfostr_pc( void *pc )
 {
-    return STATUS_UNSUCCESSFUL;
+    char buffer[256];
+
+    debugstr_pc_impl( pc, buffer, sizeof(buffer) );
+    return __wine_dbg_strdup( buffer );
 }
 
-#endif /* SO_DLLS_SUPPORTED */
+static BOOL report_native_pc_as_ntdll;
+
+static NTSTATUS is_pc_in_native_so(void *pc)
+{
+    Dl_info info;
+
+    if (!report_native_pc_as_ntdll || !dladdr( pc, &info )) return FALSE;
+
+    TRACE( "pc %p, module %s.\n", pc, debugstr_a(info.dli_fname) );
 
+    if (strstr( info.dli_fname, ".dll.so")) return FALSE;
+
+    return TRUE;
+}
 
 static const unixlib_entry_t unix_call_funcs[] =
 {
@@ -1272,6 +1305,12 @@
     unixcall_wine_server_handle_to_fd,
     unixcall_wine_spawnvp,
     system_time_precise,
+    wine_get_unix_env,
+    wine_set_unix_env,
+    unix__wine_dbg_ftrace,
+    steamclient_setup_trampolines,
+    is_pc_in_native_so,
+    debugstr_pc,
 };
 
 
@@ -1280,6 +1319,74 @@
 static NTSTATUS wow64_load_so_dll( void *args ) { return STATUS_INVALID_IMAGE_FORMAT; }
 static NTSTATUS wow64_unwind_builtin_dll( void *args ) { return STATUS_UNSUCCESSFUL; }
 
+static NTSTATUS wow64___wine_get_unix_env( void *args )
+{
+    struct
+    {
+        ULONG name;
+        ULONG val;
+        unsigned int buffer_len;
+    } const *params32 = args;
+    struct wine_get_unix_env_params params =
+    {
+        .name = ULongToPtr( params32->name ),
+        .val = ULongToPtr( params32->val ),
+        .buffer_len = params32->buffer_len,
+    };
+    return wine_get_unix_env( &params );
+}
+
+static NTSTATUS wow64___wine_set_unix_env( void *args )
+{
+    struct
+    {
+        ULONG name;
+        ULONG val;
+    } const *params32 = args;
+    struct wine_set_unix_env_params params =
+    {
+        .name = ULongToPtr( params32->name ),
+        .val = ULongToPtr( params32->val ),
+    };
+    return wine_set_unix_env( &params );
+}
+
+static NTSTATUS wow64___wine_dbg_ftrace( void *args )
+{
+    struct
+    {
+        ULONG str;
+        unsigned int len;
+        unsigned int ctx;
+    } const *params32 = args;
+    return __wine_dbg_ftrace( ULongToPtr( params32->str ), params32->len, params32->ctx );
+}
+
+static NTSTATUS wow64_steamclient_setup_trampolines( void *args )
+{
+    struct
+    {
+        ULONG src_mod;
+        ULONG tgt_mod;
+    } const *params32 = args;
+    struct steamclient_setup_trampolines_params params;
+    params.src_mod = (HMODULE)(UINT_PTR)params32->src_mod;
+    params.tgt_mod = (HMODULE)(UINT_PTR)params32->tgt_mod;
+    return steamclient_setup_trampolines( &params );
+}
+
+static NTSTATUS wow64_debugstr_pc( void *args )
+{
+    struct
+    {
+        ULONG        pc;
+        ULONG        buffer;
+        unsigned int size;
+    } const *params32 = args;
+    return debugstr_pc_impl( ULongToPtr( params32->pc ), ULongToPtr( params32->buffer ), params32->size )
+               ? STATUS_SUCCESS : STATUS_UNSUCCESSFUL;
+}
+
 const unixlib_entry_t unix_call_wow64_funcs[] =
 {
     wow64_load_so_dll,
@@ -1290,18 +1397,61 @@
     wow64_wine_server_handle_to_fd,
     wow64_wine_spawnvp,
     system_time_precise,
+    wow64___wine_get_unix_env,
+    wow64___wine_set_unix_env,
+    wow64___wine_dbg_ftrace,
+    wow64_steamclient_setup_trampolines,
+    is_pc_in_native_so,
+    wow64_debugstr_pc,
 };
 
 #endif  /* _WIN64 */
 
 
+/* check if the library is the correct architecture */
+/* only returns false for a valid library of the wrong arch */
+static int check_library_arch( int fd )
+{
+#ifdef __APPLE__
+    struct  /* Mach-O header */
+    {
+        unsigned int magic;
+        unsigned int cputype;
+    } header;
+
+    if (read( fd, &header, sizeof(header) ) != sizeof(header)) return 1;
+    if (header.magic != 0xfeedface) return 1;
+    if (sizeof(void *) == sizeof(int)) return !(header.cputype >> 24);
+    else return (header.cputype >> 24) == 1; /* CPU_ARCH_ABI64 */
+#else
+    struct  /* ELF header */
+    {
+        unsigned char magic[4];
+        unsigned char class;
+        unsigned char data;
+        unsigned char version;
+    } header;
+
+    if (read( fd, &header, sizeof(header) ) != sizeof(header)) return 1;
+    if (memcmp( header.magic, "\177ELF", 4 )) return 1;
+    if (header.version != 1 /* EV_CURRENT */) return 1;
+#ifdef WORDS_BIGENDIAN
+    if (header.data != 2 /* ELFDATA2MSB */) return 1;
+#else
+    if (header.data != 1 /* ELFDATA2LSB */) return 1;
+#endif
+    if (sizeof(void *) == sizeof(int)) return header.class == 1; /* ELFCLASS32 */
+    else return header.class == 2; /* ELFCLASS64 */
+#endif
+}
+
 static inline char *prepend( char *buffer, const char *str, size_t len )
 {
     return memcpy( buffer - len, str, len );
 }
 
 static inline char *prepend_build_dir_path( char *ptr, const char *ext, const char *arch_dir,
-                                            const char *top_dir, const char *build_dir )
+                                            const char *top_dir )
 {
     char *name = ptr;
     unsigned int namelen = strlen(name), extlen = strlen(ext);
@@ -1373,92 +1523,135 @@
 
 
 /***********************************************************************
+ *           open_builtin_so_file
+ */
+static NTSTATUS open_builtin_so_file( const char *name, OBJECT_ATTRIBUTES *attr, void **module,
+                                      SECTION_IMAGE_INFORMATION *image_info, BOOL prefer_native )
+{
+    NTSTATUS status;
+    int fd;
+
+    *module = NULL;
+    if ((fd = open( name, O_RDONLY )) == -1) return STATUS_DLL_NOT_FOUND;
+
+    if (check_library_arch( fd ))
+    {
+        struct pe_image_info info;
+
+        status = dlopen_dll( name, attr->ObjectName, module, &info, prefer_native );
+        if (!status) virtual_fill_image_information( &info, image_info );
+        else if (status != STATUS_IMAGE_ALREADY_LOADED)
+        {
+            ERR( "failed to load .so lib %s\n", debugstr_a(name) );
+            status = STATUS_PROCEDURE_NOT_FOUND;
+        }
+    }
+    else status = STATUS_NOT_SUPPORTED;
+
+    close( fd );
+    return status;
+}
+
+
+/***********************************************************************
+ *           open_main_image_so_file
+ */
+static NTSTATUS open_main_image_so_file( const char *name, UNICODE_STRING *nt_name, void **module,
+                                         SECTION_IMAGE_INFORMATION *image_info )
+{
+    struct pe_image_info pe_info;
+    NTSTATUS status;
+
+    /* remove .so extension from Windows name */
+    if (nt_name->Length > 3 * sizeof(WCHAR))
+    {
+        static const WCHAR soW[] = {'.','s','o',0};
+        WCHAR *p = nt_name->Buffer + nt_name->Length / sizeof(WCHAR);
+        if (!wcsicmp( p - 3, soW ))
+        {
+            p[-3] = 0;
+            nt_name->Length -= 3 * sizeof(WCHAR);
+        }
+    }
+    status = dlopen_dll( name, nt_name, module, &pe_info, FALSE );
+    if (!status) virtual_fill_image_information( &pe_info, image_info );
+    return status;
+}
+
+/***********************************************************************
  *           find_builtin_dll
  */
-static NTSTATUS find_builtin_dll( UNICODE_STRING *nt_name, ANSI_STRING *exp_name, void **module,
-                                  SIZE_T *size_ptr, SECTION_IMAGE_INFORMATION *image_info,
-                                  ULONG_PTR limit_low, ULONG_PTR limit_high, USHORT search_machine,
+static NTSTATUS find_builtin_dll( UNICODE_STRING *nt_name, void **module, SIZE_T *size_ptr,
+                                  SECTION_IMAGE_INFORMATION *image_info, ULONG_PTR limit_low,
+                                  ULONG_PTR limit_high, USHORT search_machine,
                                   USHORT load_machine, BOOL prefer_native, off_t offset )
 {
-    unsigned int i, pos, len, namepos = 0, maxlen = 0;
+    unsigned int i, pos, namepos, maxlen = 0;
+    unsigned int len = nt_name->Length / sizeof(WCHAR);
     char *ptr = NULL, *file, *ext = NULL;
     const char *pe_dir = get_pe_dir( search_machine );
-    const char *so_dir = get_so_dir( current_machine );
-    const char *pe_build_dir = build_dir;
     OBJECT_ATTRIBUTES attr;
     NTSTATUS status = STATUS_DLL_NOT_FOUND;
     BOOL found_image = FALSE;
+    BOOL try_so = (search_machine == current_machine && (!load_machine || load_machine == search_machine));
 
+    for (i = namepos = 0; i < len; i++)
+        if (nt_name->Buffer[i] == '/' || nt_name->Buffer[i] == '\\') namepos = i + 1;
+    len -= namepos;
+    if (!len) return STATUS_DLL_NOT_FOUND;
     InitializeObjectAttributes( &attr, nt_name, 0, 0, NULL );
 
-    if (!exp_name || !exp_name->Length)
-    {
-        len = nt_name->Length / sizeof(WCHAR);
-        for (i = 0; i < len; i++)
-            if (nt_name->Buffer[i] == '/' || nt_name->Buffer[i] == '\\') namepos = i + 1;
-        len -= namepos;
-        if (!len) return STATUS_DLL_NOT_FOUND;
-    }
-    else len = exp_name->Length;
-
-    if (build_dir)
-    {
-        if (alt_build_dir && search_machine == get_alt_machine( current_machine ))
-            pe_build_dir = alt_build_dir;
-        maxlen = max( strlen(build_dir), strlen(pe_build_dir) ) + sizeof("/programs/") + len;
-    }
+    if (build_dir) maxlen = strlen(build_dir) + sizeof("/programs/") + len;
     maxlen = max( maxlen, dll_path_maxlen + 1 ) + len + sizeof("/aarch64-windows") + sizeof(".so");
 
     if (!(file = malloc( maxlen ))) return STATUS_NO_MEMORY;
 
     pos = maxlen - len - sizeof(".so");
-    if (!exp_name || !exp_name->Length)
-    {
-        /* we don't want to depend on the current codepage here */
-        for (i = 0; i < len; i++)
-        {
-            if (nt_name->Buffer[namepos + i] > 127) goto done;
-            file[pos + i] = (char)nt_name->Buffer[namepos + i];
-        }
-    }
-    else memcpy( file + pos, exp_name->Buffer, len );
-
+    /* we don't want to depend on the current codepage here */
     for (i = 0; i < len; i++)
     {
+        if (nt_name->Buffer[namepos + i] > 127) goto done;
+        file[pos + i] = (char)nt_name->Buffer[namepos + i];
         if (file[pos + i] >= 'A' && file[pos + i] <= 'Z') file[pos + i] += 'a' - 'A';
         else if (file[pos + i] == '.') ext = file + pos + i;
     }
-    file[pos + len] = 0;
     file[--pos] = '/';
 
-    TRACE( "looking for %s for file %s\n", debugstr_a(file + pos + 1), debugstr_us(nt_name) );
-
     if (build_dir)
     {
         /* try as a dll */
-        ptr = prepend_build_dir_path( file + pos, ".dll", pe_dir, "/dlls", pe_build_dir );
+        file[pos + len + 1] = 0;
+        ptr = prepend_build_dir_path( file + pos, ".dll", pe_dir, "/dlls" );
         status = open_builtin_pe_file( ptr, &attr, module, size_ptr, image_info,
                                        limit_low, limit_high, load_machine, prefer_native, offset );
-        ptr = prepend_build_dir_path( file + pos, ".dll", "", "/dlls", build_dir );
-        if (status != STATUS_DLL_NOT_FOUND) goto done;
-        status = open_builtin_so_file( ptr, &attr, module, image_info,
-                                       search_machine, load_machine, prefer_native );
+        ptr = prepend_build_dir_path( file + pos, ".dll", "", "/dlls" );
         if (status != STATUS_DLL_NOT_FOUND) goto done;
+        if (try_so)
+        {
+            strcpy( file + pos + len + 1, ".so" );
+            status = open_builtin_so_file( ptr, &attr, module, image_info, prefer_native );
+            if (status != STATUS_DLL_NOT_FOUND) goto done;
+        }
 
         /* now as a program */
-        ptr = prepend_build_dir_path( file + pos, ".exe", pe_dir, "/programs", pe_build_dir );
+        file[pos + len + 1] = 0;
+        ptr = prepend_build_dir_path( file + pos, ".exe", pe_dir, "/programs" );
         status = open_builtin_pe_file( ptr, &attr, module, size_ptr, image_info,
                                        limit_low, limit_high, load_machine, prefer_native, offset );
-        ptr = prepend_build_dir_path( file + pos, ".exe", "", "/programs", build_dir );
-        if (status != STATUS_DLL_NOT_FOUND) goto done;
-        status = open_builtin_so_file( ptr, &attr, module, image_info,
-                                       search_machine, load_machine, prefer_native );
+        ptr = prepend_build_dir_path( file + pos, ".exe", "", "/programs" );
         if (status != STATUS_DLL_NOT_FOUND) goto done;
+        if (try_so)
+        {
+            strcpy( file + pos + len + 1, ".so" );
+            status = open_builtin_so_file( ptr, &attr, module, image_info, prefer_native );
+            if (status != STATUS_DLL_NOT_FOUND) goto done;
+        }
     }
 
     for (i = 0; dll_paths[i]; i++)
     {
         ptr = file + pos;
+        file[pos + len + 1] = 0;
         ptr = prepend( ptr, pe_dir, strlen(pe_dir) );
         ptr = prepend( ptr, dll_paths[i], strlen(dll_paths[i]) );
         status = open_builtin_pe_file( ptr, &attr, module, size_ptr, image_info, limit_low, limit_high,
@@ -1468,9 +1661,13 @@
         ptr = prepend( ptr, so_dir, strlen(so_dir) );
         ptr = prepend( ptr, dll_paths[i], strlen(dll_paths[i]) );
         if (status != STATUS_DLL_NOT_FOUND) goto done;
-        status = open_builtin_so_file( ptr, &attr, module, image_info,
-                                       search_machine, load_machine, prefer_native );
-        if (status != STATUS_DLL_NOT_FOUND) goto done;
+        if (try_so)
+        {
+            strcpy( file + pos + len + 1, ".so" );
+            status = open_builtin_so_file( ptr, &attr, module, image_info, prefer_native );
+            if (status != STATUS_DLL_NOT_FOUND) goto done;
+        }
+        file[pos + len + 1] = 0;
         ptr = prepend( file + pos, dll_paths[i], strlen(dll_paths[i]) );
         status = open_builtin_pe_file( ptr, &attr, module, size_ptr, image_info, limit_low, limit_high,
                                        load_machine, prefer_native, offset );
@@ -1480,8 +1677,9 @@
             continue;
         }
         if (status != STATUS_DLL_NOT_FOUND) goto done;
-        status = open_builtin_so_file( ptr, &attr, module, image_info,
-                                       search_machine, load_machine, prefer_native );
+        if (!try_so) continue;
+        strcpy( file + pos + len + 1, ".so" );
+        status = open_builtin_so_file( ptr, &attr, module, image_info, prefer_native );
         if (status == STATUS_NOT_SUPPORTED) found_image = TRUE;
         else if (status != STATUS_DLL_NOT_FOUND) goto done;
     }
@@ -1505,14 +1703,17 @@
  * Load the builtin dll if specified by load order configuration.
  * Return STATUS_IMAGE_ALREADY_LOADED if we should keep the native one that we have found.
  */
-NTSTATUS load_builtin( const struct pe_image_info *image_info, UNICODE_STRING *nt_name,
-                       ANSI_STRING *exp_name, USHORT machine, SECTION_IMAGE_INFORMATION *info,
-                       void **module, SIZE_T *size, ULONG_PTR limit_low, ULONG_PTR limit_high,
-                       off_t offset )
+NTSTATUS load_builtin( const struct pe_image_info *image_info, WCHAR *filename, USHORT machine,
+                       SECTION_IMAGE_INFORMATION *info, void **module, SIZE_T *size,
+                       ULONG_PTR limit_low, ULONG_PTR limit_high, off_t offset )
 {
     NTSTATUS status;
+    UNICODE_STRING nt_name;
     USHORT search_machine = image_info->machine;
-    enum loadorder loadorder = get_load_order( nt_name );
+    enum loadorder loadorder;
+
+    init_unicode_string( &nt_name, filename );
+    loadorder = get_load_order( &nt_name );
 
     if (loadorder == LO_DISABLED) return STATUS_DLL_NOT_FOUND;
 
@@ -1523,7 +1724,7 @@
     }
     else if (image_info->wine_fakedll)
     {
-        TRACE( "%s is a fake Wine dll\n", debugstr_us(nt_name) );
+        TRACE( "%s is a fake Wine dll\n", debugstr_w(filename) );
         if (loadorder == LO_NATIVE) return STATUS_DLL_NOT_FOUND;
         loadorder = LO_BUILTIN;  /* builtin with no fallback since mapping a fake dll is not useful */
     }
@@ -1537,10 +1738,10 @@
     case LO_NATIVE_BUILTIN:
         return STATUS_IMAGE_ALREADY_LOADED;
     case LO_BUILTIN:
-        return find_builtin_dll( nt_name, exp_name, module, size, info, limit_low, limit_high,
+        return find_builtin_dll( &nt_name, module, size, info, limit_low, limit_high,
                                  search_machine, machine, FALSE, offset );
     default:
-        status = find_builtin_dll( nt_name, exp_name, module, size, info, limit_low, limit_high,
+        status = find_builtin_dll( &nt_name, module, size, info, limit_low, limit_high,
                                    search_machine, machine, (loadorder == LO_DEFAULT), offset );
         if (status == STATUS_DLL_NOT_FOUND || status == STATUS_NOT_SUPPORTED)
             return STATUS_IMAGE_ALREADY_LOADED;
@@ -1605,9 +1806,10 @@
 /***********************************************************************
  *           open_main_image
  */
-static NTSTATUS open_main_image( UNICODE_STRING *nt_name, void **module, SECTION_IMAGE_INFORMATION *info,
+static NTSTATUS open_main_image( WCHAR *image, void **module, SECTION_IMAGE_INFORMATION *info,
                                  enum loadorder loadorder, USHORT machine )
 {
+    UNICODE_STRING nt_name;
     OBJECT_ATTRIBUTES attr;
     SIZE_T size = 0;
     char *unix_name;
@@ -1617,7 +1819,8 @@
 
     if (loadorder == LO_DISABLED) NtTerminateProcess( GetCurrentProcess(), STATUS_DLL_NOT_FOUND );
 
-    InitializeObjectAttributes( &attr, nt_name, OBJ_CASE_INSENSITIVE, 0, NULL );
+    init_unicode_string( &nt_name, image );
+    InitializeObjectAttributes( &attr, &nt_name, OBJ_CASE_INSENSITIVE, 0, NULL );
     if (get_nt_and_unix_names( &attr, &true_nt_name, &unix_name, FILE_OPEN, FALSE )) return STATUS_DLL_NOT_FOUND;
 
     status = open_dll_file( unix_name, &attr, &mapping );
@@ -1644,21 +1847,61 @@
 /***********************************************************************
  *           load_main_exe
  */
-NTSTATUS load_main_exe( UNICODE_STRING *nt_name, USHORT load_machine, void **module )
+NTSTATUS load_main_exe( const WCHAR *dos_name, const char *unix_name, const WCHAR *curdir,
+                        USHORT load_machine, WCHAR **image, void **module )
 {
-    enum loadorder loadorder = get_load_order( nt_name );
+    enum loadorder loadorder = LO_INVALID;
+    UNICODE_STRING nt_name;
+    WCHAR *tmp = NULL;
+    BOOL contains_path;
     unsigned int status;
     SIZE_T size;
+    struct stat st;
     USHORT search_machine;
 
-    status = open_main_image( nt_name, module, &main_image_info, loadorder, load_machine );
+    /* special case for Unix file name */
+    if (unix_name && unix_name[0] == '/' && !stat( unix_name, &st ))
+    {
+        if ((status = unix_to_nt_file_name( unix_name, image ))) goto failed;
+        init_unicode_string( &nt_name, *image );
+        loadorder = get_load_order( &nt_name );
+        status = open_main_image( *image, module, &main_image_info, loadorder, load_machine );
+        if (status != STATUS_DLL_NOT_FOUND) return status;
+        free( *image );
+    }
+
+    if (!dos_name)
+    {
+        dos_name = tmp = malloc( (strlen(unix_name) + 1) * sizeof(WCHAR) );
+        ntdll_umbstowcs( unix_name, strlen(unix_name) + 1, tmp, strlen(unix_name) + 1 );
+    }
+    contains_path = (wcschr( dos_name, '/' ) ||
+                     wcschr( dos_name, '\\' ) ||
+                     (dos_name[0] && dos_name[1] == ':'));
+
+    if ((status = get_full_path( dos_name, curdir, image ))) goto failed;
+    free( tmp );
+
+    init_unicode_string( &nt_name, *image );
+    if (loadorder == LO_INVALID) loadorder = get_load_order( &nt_name );
+
+    status = open_main_image( *image, module, &main_image_info, loadorder, load_machine );
     if (status != STATUS_DLL_NOT_FOUND) return status;
 
     /* if path is in system dir, we can load the builtin even if the file itself doesn't exist */
-    if (loadorder != LO_NATIVE && is_builtin_path( nt_name, &search_machine ))
-        status = find_builtin_dll( nt_name, NULL, module, &size, &main_image_info, 0, 0,
+    if (loadorder != LO_NATIVE && is_builtin_path( &nt_name, &search_machine ))
+    {
+        status = find_builtin_dll( &nt_name, module, &size, &main_image_info, 0, 0,
                                    search_machine, load_machine, FALSE, 0 );
-    return status;
+        if (status != STATUS_DLL_NOT_FOUND) return status;
+    }
+    if (!contains_path) return STATUS_DLL_NOT_FOUND;
+
+failed:
+    MESSAGE( "wine: failed to open %s: %x\n",
+             unix_name ? debugstr_a(unix_name) : debugstr_w(dos_name), status );
+    NtTerminateProcess( GetCurrentProcess(), status );
+    return status;  /* unreached */
 }
 
 
@@ -1667,17 +1910,18 @@
  *
  * Load start.exe as main image.
  */
-NTSTATUS load_start_exe( UNICODE_STRING *nt_name, void **module )
+NTSTATUS load_start_exe( WCHAR **image, void **module )
 {
     static const WCHAR startW[] = {'s','t','a','r','t','.','e','x','e',0};
+    UNICODE_STRING nt_name;
     unsigned int status;
     SIZE_T size;
-    WCHAR *image = malloc( sizeof("\\??\\C:\\windows\\system32\\start.exe") * sizeof(WCHAR) );
 
-    wcscpy( image, get_machine_wow64_dir( current_machine ));
-    wcscat( image, startW );
-    init_unicode_string( nt_name, image );
-    status = find_builtin_dll( nt_name, NULL, module, &size, &main_image_info, 0, 0, current_machine, 0, FALSE, 0 );
+    *image = malloc( sizeof("\\??\\C:\\windows\\system32\\start.exe") * sizeof(WCHAR) );
+    wcscpy( *image, get_machine_wow64_dir( current_machine ));
+    wcscat( *image, startW );
+    init_unicode_string( &nt_name, *image );
+    status = find_builtin_dll( &nt_name, module, &size, &main_image_info, 0, 0, current_machine, 0, FALSE, 0 );
     if (!NT_SUCCESS(status))
     {
         MESSAGE( "wine: failed to load start.exe: %x\n", status );
@@ -1686,53 +1930,6 @@
     return status;
 }
 
-static ULONG_PTR find_ordinal_export( HMODULE module, const IMAGE_EXPORT_DIRECTORY *exports, DWORD ordinal )
-{
-    const DWORD *functions = (const DWORD *)((BYTE *)module + exports->AddressOfFunctions);
-
-    if (ordinal >= exports->NumberOfFunctions) return 0;
-    if (!functions[ordinal]) return 0;
-    return (ULONG_PTR)module + functions[ordinal];
-}
-
-static ULONG_PTR find_named_export( HMODULE module, const IMAGE_EXPORT_DIRECTORY *exports,
-                                    const char *name )
-{
-    const WORD *ordinals = (const WORD *)((BYTE *)module + exports->AddressOfNameOrdinals);
-    const DWORD *names = (const DWORD *)((BYTE *)module + exports->AddressOfNames);
-    int min = 0, max = exports->NumberOfNames - 1;
-
-    while (min <= max)
-    {
-        int res, pos = (min + max) / 2;
-        char *ename = (char *)module + names[pos];
-        if (!(res = strcmp( ename, name ))) return find_ordinal_export( module, exports, ordinals[pos] );
-        if (res > 0) max = pos - 1;
-        else min = pos + 1;
-    }
-    return 0;
-}
-
-static inline void *get_rva( void *module, ULONG_PTR addr )
-{
-    return (BYTE *)module + addr;
-}
-
-static const void *get_module_data_dir( HMODULE module, ULONG dir, ULONG *size )
-{
-    const IMAGE_NT_HEADERS *nt = get_rva( module, ((IMAGE_DOS_HEADER *)module)->e_lfanew );
-    const IMAGE_DATA_DIRECTORY *data;
-
-    if (nt->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR64_MAGIC)
-        data = &((const IMAGE_NT_HEADERS64 *)nt)->OptionalHeader.DataDirectory[dir];
-    else if (nt->OptionalHeader.Magic == IMAGE_NT_OPTIONAL_HDR32_MAGIC)
-        data = &((const IMAGE_NT_HEADERS32 *)nt)->OptionalHeader.DataDirectory[dir];
-    else
-        return NULL;
-    if (!data->VirtualAddress || !data->Size) return NULL;
-    if (size) *size = data->Size;
-    return get_rva( module, data->VirtualAddress );
-}
 
 /***********************************************************************
  *           load_ntdll_functions
@@ -1896,8 +2093,8 @@
     if (status == STATUS_DLL_NOT_FOUND)
     {
         free( name );
-        asprintf( &name, "%s/ntdll.dll%c.so", ntdll_dir, 0 );
-        status = open_builtin_so_file( name, &attr, &module, &info, machine, 0, FALSE );
+        asprintf( &name, "%s/ntdll.dll.so", ntdll_dir );
+        status = open_builtin_so_file( name, &attr, &module, &info, FALSE );
     }
     if (status == STATUS_IMAGE_NOT_AT_BASE) status = virtual_relocate_module( module );
     if (status) fatal_error( "failed to load %s error %x\n", name, status );
@@ -1998,7 +2195,7 @@
     wcscpy( path, wow64_dir );
     wcscat( path, ntdllW );
     init_unicode_string( &nt_name, path );
-    status = find_builtin_dll( &nt_name, NULL, &module, &size, &info, 0, 0, machine, 0, FALSE, 0 );
+    status = find_builtin_dll( &nt_name, &module, &size, &info, 0, 0, machine, 0, FALSE, 0 );
     if (status == STATUS_IMAGE_NOT_AT_BASE) status = virtual_relocate_module( module );
     if (status) fatal_error( "failed to load %s error %x\n", debugstr_w(path), status );
     load_ntdll_wow64_functions( module );
@@ -2035,6 +2232,162 @@
     return 0;
 }
 
+BOOL ac_odyssey;
+BOOL fsync_simulate_sched_quantum;
+BOOL alert_simulate_sched_quantum;
+BOOL fsync_yield_to_waiters;
+BOOL fsync_help_simulated_pulse;
+BOOL localsystem_sid;
+BOOL simulate_writecopy;
+BOOL wine_allocs_2g_limit;
+SIZE_T kernel_stack_size = 0x100000;
+long long ram_reporting_bias;
+char *release_reserved_memory_low_bound;
+
+static void hacks_init(void)
+{
+    const char *sgi = getenv( "SteamGameId" );
+    const char *env_str;
+    if ((env_str = getenv("WINE_RAM_REPORTING_BIAS")))
+    {
+        ram_reporting_bias = atoll(env_str) * 1024 * 1024;
+        ERR( "HACK: ram_reporting_bias %lldMB.\n", ram_reporting_bias / (1024 * 1024) );
+    }
+
+    env_str = getenv("WINE_SIMULATE_ASYNC_READ");
+    if (env_str)
+        ac_odyssey = !!atoi(env_str);
+    else if (main_argc > 1 && (strstr(main_argv[1], "ACOdyssey.exe") || strstr(main_argv[1], "ImmortalsFenyxRising.exe")))
+        ac_odyssey = TRUE;
+
+    if (ac_odyssey)
+        ERR("HACK: AC Odyssey sync tweak on.\n");
+
+    env_str = getenv("WINE_FSYNC_SIMULATE_SCHED_QUANTUM");
+    if (env_str)
+        fsync_simulate_sched_quantum = !!atoi(env_str);
+    else if (main_argc > 1)
+    {
+        fsync_simulate_sched_quantum = !!strstr(main_argv[1], "Ubisoft Game Launcher\\upc.exe");
+        fsync_simulate_sched_quantum = fsync_simulate_sched_quantum || !!strstr(main_argv[1], "PlanetZoo.exe");
+        fsync_simulate_sched_quantum = fsync_simulate_sched_quantum || !!strstr(main_argv[1], "GTA5.exe");
+    }
+    if (fsync_simulate_sched_quantum)
+        ERR("HACK: Simulating sched quantum in fsync.\n");
+
+    env_str = getenv("WINE_ALERT_SIMULATE_SCHED_QUANTUM");
+    if (env_str)
+        alert_simulate_sched_quantum = !!atoi(env_str);
+    else if (main_argc > 1)
+    {
+        alert_simulate_sched_quantum = !!strstr(main_argv[1], "GTA5.exe");
+        alert_simulate_sched_quantum = alert_simulate_sched_quantum || !!strstr(main_argv[1], "MarySkelter2.exe");
+        alert_simulate_sched_quantum = alert_simulate_sched_quantum || !!strstr(main_argv[1], "MarySkelterFinale.exe");
+        alert_simulate_sched_quantum = alert_simulate_sched_quantum || !!strstr(main_argv[1], "Application.exe");
+        alert_simulate_sched_quantum = alert_simulate_sched_quantum || !!strstr(main_argv[1], "DeathEndReQuest2.exe");
+        alert_simulate_sched_quantum = alert_simulate_sched_quantum || !!strstr(main_argv[1], "v2r.exe");
+        alert_simulate_sched_quantum = alert_simulate_sched_quantum || !!strstr(main_argv[1], "NeptuniaVirtualStars.exe");
+        alert_simulate_sched_quantum = alert_simulate_sched_quantum || !!strstr(main_argv[1], "DragonStarVarnir.exe");
+    }
+    if (alert_simulate_sched_quantum)
+        ERR("HACK: Simulating sched quantum in NtWaitForAlertByThreadId.\n");
+
+    env_str = getenv("WINE_FSYNC_YIELD_TO_WAITERS");
+    if (env_str)
+        fsync_yield_to_waiters = !!atoi(env_str);
+    else if (sgi) fsync_yield_to_waiters = !strcmp(sgi, "292120") || !strcmp(sgi, "345350") || !strcmp(sgi, "292140")
+                                           || !strcmp(sgi, "33460");
+    if (fsync_yield_to_waiters)
+        ERR("HACK: fsync: yield to waiters.\n");
+
+    env_str = getenv("WINE_FSYNC_HELP_SIMULATED_PULSE");
+    if (env_str)
+        fsync_help_simulated_pulse = !!atoi(env_str);
+    else if (sgi) fsync_help_simulated_pulse =
+        !strcmp(sgi, "460870")
+        || !strcmp(sgi, "1486920")
+        || !strcmp(sgi, "438490");
+
+    if (fsync_help_simulated_pulse)
+        ERR("HACK: fsync: helping simulated pulse event.\n");
+
+    switch (sgi ? atoi( sgi ) : -1)
+    {
+    case 25700: /* Madballs in Babo: Invasion */
+    case 50130: /* Mafia II */
+    case 202990: /* CoD Black Ops II Multiplayer */
+    case 212910: /* CoD Black Ops II Zombies */
+    case 247910: /* Sniper Elite: Nazi Zombie Army 2 */
+    case 227100: /* Sniper Elite: Nazi Zombie Army */
+        setenv( "WINESTEAMNOEXEC", "1", 0 );
+        break;
+    }
+
+    env_str = getenv("WINE_UNIX_PC_AS_NTDLL");
+    if (env_str)  report_native_pc_as_ntdll = atoi(env_str);
+    else if (sgi) report_native_pc_as_ntdll = !strcmp(sgi, "700330");
+
+    env_str = getenv("WINE_SIMULATE_WRITECOPY");
+    if (env_str) simulate_writecopy = atoi(env_str);
+    else if (main_argc > 1 &&
+                          (strstr(main_argv[1], "UplayWebCore.exe")
+                           || (strstr(main_argv[1], "Battle.net.exe"))))
+        simulate_writecopy = TRUE;
+    else if (sgi) simulate_writecopy = !strcmp(sgi, "1608730") /* Dawn of Corruption */
+                                       || !strcmp(sgi, "1680700") /* Purgo box */
+                                       || !strcmp(sgi, "2095300") /* Breakout 13 */
+                                       || !strcmp(sgi, "2053940") /* Idol Hands 2 */
+                                       || !strcmp(sgi, "391150") /* Red Tie Runner */
+                                       || !strcmp(sgi, "2152990") /* Dinogen Online */
+                                       || !strcmp(sgi, "2176450") /* Mr. Hopp's Playhouse 3 */
+                                       || !strcmp(sgi, "2329630") /* Lovey-Dovey Lockdown */
+                                       || !strcmp(sgi, "2209020") /* Gemstones */
+                                       || !strcmp(sgi, "223750") /* DCS World */
+                                       || !strcmp(sgi, "2361360"); /* Hentai Maid Memories */
+
+    if (sgi) wine_allocs_2g_limit = !strcmp(sgi, "359870");
+    if (wine_allocs_2g_limit) ERR("Allocation 2g limit enabled.\n");
+
+    if (main_argc > 1 && strstr(main_argv[1], "MicrosoftEdgeUpdate.exe"))
+    {
+        ERR("HACK: reporting LocalSystem account SID.\n");
+        localsystem_sid = TRUE;
+        return;
+    }
+
+    if ((env_str = getenv( "WINE_KERNEL_STACK_SIZE" )))
+        kernel_stack_size = atoll( env_str ) * 1024;
+    else if (sgi && !strcmp( sgi, "702700" ))
+        kernel_stack_size = 200 * 1024;
+    if (kernel_stack_size != 0x100000)
+        ERR( "HACK: setting kernel_stack_size to %luKB.\n", (long)(kernel_stack_size / 1024) );
+
+    if (sgi && (0
+        || !strcmp(sgi, "1364780") || !strcmp(sgi, "1952120") || !strcmp(sgi, "2154900") /* Street Fighter 6 */
+        || !strcmp(sgi, "1740720") /* Have a Nice Death  */
+    ))
+    {
+        ERR("HACK: setting WINE_ENABLE_GST_LIVE_LATENCY.\n");
+        setenv("WINE_ENABLE_GST_LIVE_LATENCY", "1", 0);
+    }
+
+    if (sgi && !strcmp(sgi, "2379390"))
+    {
+        ERR("HACK: setting vk_x11_override_min_image_count, vk_x11_strict_image_count.\n");
+        setenv("vk_x11_override_min_image_count", "2", 0);
+        setenv("vk_x11_strict_image_count", "true", 0);
+    }
+
+#ifndef __x86_64__
+    if ((env_str = getenv( "WINE_RES_MEM_LOW_BOUND" )))
+        release_reserved_memory_low_bound = (void *)strtol( env_str, NULL, 0x10 );
+    else if (sgi && (
+                        !strcmp( sgi, "518920" )
+                    ))
+        release_reserved_memory_low_bound = (void *)0x00200000;
+#endif
+}
+
 /***********************************************************************
  *           start_main_thread
  */
@@ -2043,11 +2396,20 @@
     TEB *teb = virtual_alloc_first_teb();
 
     signal_init_threading();
+    signal_alloc_thread( teb );
     dbg_init();
     startup_info_size = server_init_process();
+    hacks_init();
+    fsync_init();
+    esync_init();
     virtual_map_user_shared_data();
     init_cpu_info();
     init_files();
+
+#if defined(__x86_64__) || defined(__i386__)
+    set_thread_teb( teb );
+#endif
+
     init_startup_info();
     *(ULONG_PTR *)&peb->CloudFileFlags = get_image_address();
     set_load_order_app_name( main_wargv[0] );
@@ -2203,6 +2565,9 @@
 
     pthread_attr_init( &attr );
     pthread_attr_setdetachstate( &attr, PTHREAD_CREATE_JOINABLE );
+    /* Use the same QoS class as the process main thread (user-interactive). */
+    if (&pthread_attr_set_qos_class_np)
+        pthread_attr_set_qos_class_np( &attr, QOS_CLASS_USER_INTERACTIVE, 0 );
     if (pthread_create( &thread, &attr, apple_wine_thread, NULL )) exit(1);
     pthread_attr_destroy( &attr );
 }
@@ -2285,25 +2650,6 @@
     return 1;
 }
 
-#elif defined(__APPLE__)
-
-static int pre_exec(void)
-{
-    if (build_dir)
-    {
-        char *path = getenv( "DYLD_LIBRARY_PATH" );
-        if (path) asprintf( &path, "%s/dlls/ntdll:%s/dlls/win32u:%s", build_dir, build_dir, path );
-        else asprintf( &path, "%s/dlls/ntdll:%s/dlls/win32u", build_dir, build_dir );
-        setenv( "DYLD_LIBRARY_PATH", path, 1 );
-        return 1;
-    }
-#ifdef HAVE_WINE_PRELOADER
-    return 1;
-#else
-    return 0;
-#endif
-}
-
 #else
 
 static int pre_exec(void)
@@ -2318,33 +2664,6 @@
 #endif
 
 
-static void reexec_loader( int argc, char *argv[], char *extra_arg )
-{
-    WORD machine = current_machine;
-    char **new_argv;
-
-    /* have to exec if we have a preloader, or an argument, or if we are the initial wrapper */
-    if (!pre_exec() && !extra_arg && dlsym( RTLD_DEFAULT, "wine_main_preload_info" )) return;
-
-    if (extra_arg)
-    {
-        new_argv = malloc( (argc + 3) * sizeof(*argv) );
-        memcpy( new_argv + 3, argv + 1, argc * sizeof(*argv) );
-        new_argv[2] = extra_arg;
-    }
-    else
-    {
-        new_argv = malloc( (argc + 2) * sizeof(*argv) );
-        memcpy( new_argv + 2, argv + 1, argc * sizeof(*argv) );
-    }
-
-    /* default to 32-bit loader to support 32-bit prefixes */
-    if (machine == IMAGE_FILE_MACHINE_AMD64) machine = IMAGE_FILE_MACHINE_I386;
-
-    loader_exec( new_argv, machine );
-    fatal_error( "could not exec the wine loader\n" );
-}
-
 /***********************************************************************
  *           check_command_line
  *
@@ -2352,37 +2671,11 @@
  */
 static void check_command_line( int argc, char *argv[] )
 {
-    char *basename;
     static const char usage[] =
         "Usage: wine PROGRAM [ARGUMENTS...]   Run the specified program\n"
         "       wine --help                   Display this help and exit\n"
         "       wine --version                Output version information and exit";
 
-    if ((basename = strrchr( argv[0], '/' ))) basename++;
-    else basename = argv[0];
-
-    if (strcmp( basename, "wine" )) /* check if there's a builtin exe corresponding to the base name */
-    {
-        const char *pe_dir = get_pe_dir( current_machine );
-        char *exe;
-
-        if (build_dir)
-        {
-            asprintf( &exe, "%s/programs/%s%s/%s.exe", build_dir, basename, pe_dir, basename );
-            if (!access( exe, R_OK )) reexec_loader( argc, argv, basename );
-            free( exe );
-        }
-        else
-        {
-            for (int i = 0; dll_paths[i]; i++)
-            {
-                asprintf( &exe, "%s%s/%s.exe", dll_paths[i], pe_dir, basename );
-                if (!access( exe, R_OK )) reexec_loader( argc, argv, basename );
-                free( exe );
-            }
-        }
-    }
-
     if (argc <= 1)
     {
         fprintf( stderr, "%s\n", usage );
@@ -2398,8 +2691,6 @@
         printf( "%s\n", wine_build );
         exit(0);
     }
-
-    reexec_loader( argc, argv, NULL );
 }
 
 
@@ -2413,8 +2704,9 @@
     main_argc = argc;
     main_argv = argv;
 
-    init_paths();
-    if (!getenv( "WINELOADERNOEXEC" ))
+    init_paths( argv );
+
+    if (!getenv( "WINELOADERNOEXEC" ))  /* first time around */
     {
         check_command_line( argc, argv );
         if (pre_exec())
@@ -2428,8 +2720,6 @@
             fatal_error( "could not exec the wine loader\n" );
         }
     }
-    else if (argc <= 1) check_command_line( argc, argv );
-    unsetenv( "WINELOADERNOEXEC" );
 
 #ifdef RLIMIT_NOFILE
     set_max_limit( RLIMIT_NOFILE );
