diff --git a/dlls/opengl32/tests/opengl.c b/dlls/opengl32/tests/opengl.c
--- a/dlls/opengl32/tests/opengl.c
+++ b/dlls/opengl32/tests/opengl.c
@@ -18,54 +18,19 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#include <stdarg.h>
-#include <stddef.h>
-
-#include "ntstatus.h"
-#define WIN32_NO_STATUS
-#include "windef.h"
-#include "winbase.h"
-#include "winerror.h"
-#include "wingdi.h"
-#include "winuser.h"
-#include "winternl.h"
-#include "ddk/d3dkmthk.h"
-
+#include <windows.h>
+#include <wingdi.h>
 #include "wine/test.h"
 #include "wine/wgl.h"
 
 #define MAX_FORMATS 256
 
-static const char *debugstr_ok( const char *cond )
-{
-    int c, n = 0;
-    /* skip possible casts */
-    while ((c = *cond++))
-    {
-        if (c == '(') n++;
-        if (!n) break;
-        if (c == ')') n--;
-    }
-    if (!strchr( cond - 1, '(' )) return wine_dbg_sprintf( "got %s", cond - 1 );
-    return wine_dbg_sprintf( "%.*s returned", (int)strcspn( cond - 1, "( " ), cond - 1 );
-}
-
-#define ok_ex( r, op, e, t, f, ... )                                                               \
-    do                                                                                             \
-    {                                                                                              \
-        t v = (r);                                                                                 \
-        ok( v op (e), "%s " f "\n", debugstr_ok( #r ), v, ##__VA_ARGS__ );                         \
-    } while (0)
-#define ok_ret( e, r )      ok_ex( r, ==, e, UINT, "%#x" )
-
-static NTSTATUS (WINAPI *pD3DKMTCreateDCFromMemory)( D3DKMT_CREATEDCFROMMEMORY *desc );
-static NTSTATUS (WINAPI *pD3DKMTDestroyDCFromMemory)( const D3DKMT_DESTROYDCFROMMEMORY *desc );
-
 /* WGL_ARB_create_context */
 static HGLRC (WINAPI *pwglCreateContextAttribsARB)(HDC hDC, HGLRC hShareContext, const int *attribList);
 
 /* WGL_ARB_extensions_string */
 static const char* (WINAPI *pwglGetExtensionsStringARB)(HDC);
+static int (WINAPI *pwglReleasePbufferDCARB)(HPBUFFERARB, HDC);
 
 /* WGL_ARB_make_current_read */
 static BOOL (WINAPI *pwglMakeContextCurrentARB)(HDC hdraw, HDC hread, HGLRC hglrc);
@@ -77,15 +42,7 @@
 
 /* WGL_ARB_pbuffer */
 static HPBUFFERARB (WINAPI *pwglCreatePbufferARB)(HDC, int, int, int, const int *);
-static BOOL (WINAPI *pwglDestroyPbufferARB)(HPBUFFERARB);
 static HDC (WINAPI *pwglGetPbufferDCARB)(HPBUFFERARB);
-static int (WINAPI *pwglReleasePbufferDCARB)(HPBUFFERARB, HDC);
-static BOOL (WINAPI *pwglQueryPbufferARB)(HPBUFFERARB,int,int*);
-
-/* WGL_ARB_render_texture */
-static BOOL (WINAPI *pwglBindTexImageARB)(HPBUFFERARB,int);
-static BOOL (WINAPI *pwglReleaseTexImageARB)(HPBUFFERARB,int);
-static BOOL (WINAPI *pwglSetPbufferAttribARB)(HPBUFFERARB,const int*);
 
 /* WGL_EXT_swap_control */
 static BOOL (WINAPI *pwglSwapIntervalEXT)(int interval);
@@ -99,27 +56,6 @@
 /* GL_ARB_framebuffer_object */
 static void (WINAPI *pglBindFramebuffer)(GLenum target, GLuint framebuffer);
 static GLenum (WINAPI *pglCheckFramebufferStatus)(GLenum target);
-
-static PFN_glBindBuffer pglBindBuffer;
-static PFN_glBufferData pglBufferData;
-static PFN_glBufferStorage pglBufferStorage;
-static PFN_glCopyBufferSubData pglCopyBufferSubData;
-static PFN_glCopyNamedBufferSubData pglCopyNamedBufferSubData;
-static PFN_glCreateBuffers pglCreateBuffers;
-static PFN_glDeleteBuffers pglDeleteBuffers;
-static PFN_glDeleteSync pglDeleteSync;
-static PFN_glFlushMappedBufferRange pglFlushMappedBufferRange;
-static PFN_glFlushMappedNamedBufferRange pglFlushMappedNamedBufferRange;
-static PFN_glGenBuffers pglGenBuffers;
-static PFN_glIsSync pglIsSync;
-static PFN_glMapBuffer pglMapBuffer;
-static PFN_glMapBufferRange pglMapBufferRange;
-static PFN_glMapNamedBuffer pglMapNamedBuffer;
-static PFN_glMapNamedBufferRange pglMapNamedBufferRange;
-static PFN_glNamedBufferData pglNamedBufferData;
-static PFN_glNamedBufferStorage pglNamedBufferStorage;
-static PFN_glUnmapBuffer pglUnmapBuffer;
-static PFN_glUnmapNamedBuffer pglUnmapNamedBuffer;
 
 static const char* wgl_extensions = NULL;
 
@@ -163,15 +99,8 @@
 
     /* WGL_ARB_pbuffer */
     GET_PROC(wglCreatePbufferARB)
-    GET_PROC(wglDestroyPbufferARB)
     GET_PROC(wglGetPbufferDCARB)
     GET_PROC(wglReleasePbufferDCARB)
-    GET_PROC(wglQueryPbufferARB)
-
-    /* WGL_ARB_render_texture */
-    GET_PROC(wglBindTexImageARB)
-    GET_PROC(wglReleaseTexImageARB)
-    GET_PROC(wglSetPbufferAttribARB)
 
     /* WGL_EXT_swap_control */
     GET_PROC(wglSwapIntervalEXT)
@@ -185,27 +114,6 @@
     /* GL_ARB_framebuffer_object */
     GET_PROC(glBindFramebuffer)
     GET_PROC(glCheckFramebufferStatus)
-
-    GET_PROC(glBindBuffer)
-    GET_PROC(glBufferData)
-    GET_PROC(glBufferStorage)
-    GET_PROC(glCopyBufferSubData)
-    GET_PROC(glCopyNamedBufferSubData)
-    GET_PROC(glCreateBuffers)
-    GET_PROC(glDeleteBuffers)
-    GET_PROC(glDeleteSync)
-    GET_PROC(glFlushMappedBufferRange)
-    GET_PROC(glFlushMappedNamedBufferRange)
-    GET_PROC(glGenBuffers)
-    GET_PROC(glIsSync)
-    GET_PROC(glMapBuffer)
-    GET_PROC(glMapBufferRange)
-    GET_PROC(glMapNamedBuffer)
-    GET_PROC(glMapNamedBufferRange)
-    GET_PROC(glNamedBufferData)
-    GET_PROC(glNamedBufferStorage)
-    GET_PROC(glUnmapBuffer)
-    GET_PROC(glUnmapNamedBuffer)
 
 #undef GET_PROC
 }
@@ -237,558 +145,104 @@
     return FALSE;
 }
 
-static void test_pbuffers( HDC old_hdc )
-{
-    int attribs[32] = { WGL_DRAW_TO_PBUFFER_ARB, 1, 0 };
-    int formats[MAX_FORMATS], pbuffer_attribs[15] = {0};
-    unsigned int i, count, onscreen;
-    unsigned int pixels[16 * 16];
-    HDC hdc, pbuffer_dc, tmp_dc;
-    HPBUFFERARB pbuffer;
-    HGLRC rc, old_rc;
-    int res, value;
-    GLuint texture;
-    HWND hwnd;
-    BOOL ret;
-
-    old_rc = wglGetCurrentContext();
-
-    hwnd = CreateWindowW( L"static", NULL, WS_POPUP, 10, 10, 200, 200, NULL, NULL, NULL, NULL );
-    ok( !!hwnd, "CreateWindow failed, error %lu\n", GetLastError() );
-    hdc = GetDC( hwnd );
-    ok( !!hdc, "GetDC failed, error %lu\n", GetLastError() );
-
-    onscreen = DescribePixelFormat( hdc, 0, 0, NULL );
-    attribs[0] = WGL_DRAW_TO_WINDOW_ARB; attribs[1] = 1;
-    attribs[2] = WGL_COLOR_BITS_ARB; attribs[3] = 32;
-    attribs[4] = WGL_PIXEL_TYPE_ARB; attribs[5] = WGL_TYPE_RGBA_ARB;
-    res = pwglChoosePixelFormatARB( hdc, attribs, NULL, MAX_FORMATS, formats, &count );
-    ok( res > 0, "got %d\n", res );
-    ret = SetPixelFormat( hdc, formats[0], NULL );
-    ok( ret == 1, "got %u\n", ret );
-
-    attribs[0] = WGL_DRAW_TO_PBUFFER_ARB; attribs[1] = 1;
-    attribs[2] = WGL_COLOR_BITS_ARB; attribs[3] = 32;
-    attribs[4] = WGL_PIXEL_TYPE_ARB; attribs[5] = WGL_TYPE_RGBA_ARB;
-    res = pwglChoosePixelFormatARB( hdc, attribs, NULL, MAX_FORMATS, formats, &count );
-    ok( res > 0, "got %d\n", res );
-    if (count > MAX_FORMATS) count = MAX_FORMATS;
-
-    wglMakeCurrent( 0, 0 );
-
-    SetLastError( 0xdeadbeef );
-    pbuffer = pwglCreatePbufferARB( hdc, 0, 100, 100, pbuffer_attribs );
-    ok( !pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_PIXEL_FORMAT, "got %lu\n", GetLastError() );
-    if (pbuffer) pwglDestroyPbufferARB( pbuffer );
-    SetLastError( 0xdeadbeef );
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 0, 100, pbuffer_attribs );
-    ok( !pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA, "got %lu\n", GetLastError() );
-    if (pbuffer) pwglDestroyPbufferARB( pbuffer );
-    SetLastError( 0xdeadbeef );
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], -1, 100, pbuffer_attribs );
-    ok( !pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA, "got %lu\n", GetLastError() );
-    SetLastError( 0xdeadbeef );
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 100, 0, pbuffer_attribs );
-    ok( !pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA, "got %lu\n", GetLastError() );
-    if (pbuffer) pwglDestroyPbufferARB( pbuffer );
-    SetLastError( 0xdeadbeef );
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 100, -1, pbuffer_attribs );
-    ok( !pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA, "got %#lx\n", GetLastError() );
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 100, 100, NULL );
-    ok( !!pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-    pwglDestroyPbufferARB( pbuffer );
-
-    for (i = 0; i < count; i++)
-    {
-        winetest_push_context( "%u", formats[i] );
-        pbuffer = pwglCreatePbufferARB( hdc, formats[i], 640, 480, pbuffer_attribs );
-        ok( !!pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-        pbuffer_dc = pwglGetPbufferDCARB( pbuffer );
-        ok( pbuffer_dc != hdc, "got %p\n", pbuffer_dc );
-        res = GetPixelFormat( pbuffer_dc );
-        ret = pwglReleasePbufferDCARB( pbuffer, pbuffer_dc );
-        ok( ret == 1, "got %u\n", ret );
-        if (formats[i] > onscreen) ok( res == 1, "got format %d\n", res );
-        else ok( res == formats[i] || broken( res == 1 ) /* AMD sometimes */, "got format %d\n", res );
-        ret = pwglDestroyPbufferARB( pbuffer );
-        ok( ret == 1, "got %u\n", ret );
-        winetest_pop_context();
-    }
-
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 640, 480, pbuffer_attribs );
-    ok( !!pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-
-    pbuffer_dc = pwglGetPbufferDCARB( pbuffer );
-    ok( pbuffer_dc != hdc, "got %p\n", pbuffer_dc );
-
-    /* wglGetPbufferDCARB returns the same DC every time */
-    tmp_dc = pwglGetPbufferDCARB( pbuffer );
-    ok( tmp_dc == pbuffer_dc, "got %p\n", tmp_dc );
-
-    /* releasing the wrong DC returns an error */
-    SetLastError( 0xdeadbeef );
-    ret = pwglReleasePbufferDCARB( pbuffer, hdc );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_DC_NOT_FOUND, "got %#lx\n", GetLastError() );
-
-    ret = pwglReleasePbufferDCARB( pbuffer, pbuffer_dc );
-    ok( ret == 1, "got %u\n", ret );
-    /* releasing the DC more than once may return an error */
-    SetLastError( 0xdeadbeef );
-    ret = pwglReleasePbufferDCARB( pbuffer, pbuffer_dc );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-    if (!ret) todo_wine ok( (GetLastError() & 0xffff) == ERROR_DC_NOT_FOUND, "got %#lx\n", GetLastError() );
-    SetLastError( 0xdeadbeef );
-    ret = pwglReleasePbufferDCARB( pbuffer, pbuffer_dc );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-    if (!ret) todo_wine ok( (GetLastError() & 0xffff) == ERROR_DC_NOT_FOUND, "got %#lx\n", GetLastError() );
-
-    tmp_dc = pwglGetPbufferDCARB( pbuffer );
-    if (!ret) ok( tmp_dc != pbuffer_dc, "got %p\n", tmp_dc );
-    else ok( tmp_dc == pbuffer_dc, "got %p\n", tmp_dc );
-    ret = pwglReleasePbufferDCARB( pbuffer, tmp_dc );
-    ok( ret == 1, "got %u\n", ret );
-
-    SetLastError( 0xdeadbeef );
-    ret = pwglQueryPbufferARB( NULL, WGL_PBUFFER_WIDTH_ARB, &value );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_HANDLE, "got %#lx\n", GetLastError() );
-    SetLastError( 0xdeadbeef );
-    ret = pwglQueryPbufferARB( pbuffer, 0, &value );
-    todo_wine ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA, "got %#lx\n", GetLastError() );
-    SetLastError( 0xdeadbeef );
-    ret = pwglQueryPbufferARB( pbuffer, 0xdeadbeef, &value );
-    todo_wine ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA, "got %#lx\n", GetLastError() );
-
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_WIDTH_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 0 || value == 640, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_HEIGHT_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 0 || value == 480, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_LOST_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 0, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_TEXTURE_FORMAT_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == WGL_NO_TEXTURE_ARB || broken(value == 0xdeadbeef) /* AMD */, "got %#x\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_TEXTURE_TARGET_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == WGL_NO_TEXTURE_ARB || broken(value == 0xdeadbeef) /* AMD */, "got %#x\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_MIPMAP_TEXTURE_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 0 || broken(value > 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_MIPMAP_LEVEL_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 0 || broken(value > 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_CUBE_MAP_FACE_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == WGL_TEXTURE_CUBE_MAP_POSITIVE_X_ARB || broken(value == 0xdeadbeef), "got %#x\n", value );
-
-    pbuffer_attribs[0] = WGL_PBUFFER_WIDTH_ARB;
-    pbuffer_attribs[1] = 50;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_PBUFFER_HEIGHT_ARB;
-    pbuffer_attribs[1] = 50;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_PBUFFER_LOST_ARB;
-    pbuffer_attribs[1] = 0;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_TEXTURE_FORMAT_ARB;
-    pbuffer_attribs[1] = WGL_TEXTURE_RGBA_ARB;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_TEXTURE_TARGET_ARB;
-    pbuffer_attribs[1] = WGL_TEXTURE_2D_ARB;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_MIPMAP_TEXTURE_ARB;
-    pbuffer_attribs[1] = 1;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_MIPMAP_LEVEL_ARB;
-    pbuffer_attribs[1] = 1;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0 || broken(ret == 1) /* AMD */, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_CUBE_MAP_FACE_ARB;
-    pbuffer_attribs[1] = WGL_TEXTURE_CUBE_MAP_POSITIVE_X_ARB;
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    todo_wine ok( ret == 1, "got %u\n", ret );
-
-    SetLastError( 0xdeadbeef );
-    ret = pwglDestroyPbufferARB( pbuffer );
-    ok( ret == 1, "got %u\n", ret );
-    ok( GetLastError() == 0xdeadbeef, "got %#lx\n", GetLastError() );
-    /* destroying the pbuffer multiple times is an error */
-    SetLastError( 0xdeadbeef );
-    ret = pwglDestroyPbufferARB( pbuffer );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_HANDLE, "got %#lx\n", GetLastError() );
-
-    if (!winetest_platform_is_wine) /* triggers a BadAlloc */
-    {
-    pbuffer_attribs[0] = WGL_PBUFFER_LARGEST_ARB;
-    pbuffer_attribs[1] = 1;
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 65535, 65535, pbuffer_attribs );
-    ok( !!pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_WIDTH_ARB, &value );
-    ok( ret == 1 || ret == 0, "got %u\n", ret );
-    ok( value > 0 && value < 65535, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_HEIGHT_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value > 0 && value < 65535, "got %u\n", value );
-    pwglDestroyPbufferARB( pbuffer );
-
-    pbuffer_attribs[0] = WGL_PBUFFER_LARGEST_ARB;
-    pbuffer_attribs[1] = 0;
-    SetLastError( 0xdeadbeef );
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 65535, 65535, pbuffer_attribs );
-    ok( !pbuffer || broken(!!pbuffer) /* AMD */, "wglCreatePbufferARB returned %p\n", pbuffer );
-    ok( (GetLastError() & 0xffff) == ERROR_NO_SYSTEM_RESOURCES || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    if (pbuffer) pwglDestroyPbufferARB( pbuffer );
-    }
-
-    pbuffer_attribs[0] = WGL_TEXTURE_FORMAT_ARB;
-    pbuffer_attribs[1] = WGL_TEXTURE_RGB_ARB;
-    pbuffer_attribs[2] = WGL_TEXTURE_TARGET_ARB;
-    pbuffer_attribs[3] = WGL_TEXTURE_CUBE_MAP_ARB;
-    pbuffer_attribs[4] = WGL_MIPMAP_TEXTURE_ARB;
-    pbuffer_attribs[5] = 4;
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 512, 512, pbuffer_attribs );
-    ok( !!pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_WIDTH_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 512 || broken(value == 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_HEIGHT_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 512 || broken(value == 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_LOST_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 0, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_TEXTURE_FORMAT_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == WGL_TEXTURE_RGB_ARB || broken(value == 0xdeadbeef) /* AMD */, "got %#x\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_TEXTURE_TARGET_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == WGL_TEXTURE_CUBE_MAP_ARB || broken(value == 0xdeadbeef) /* AMD */, "got %#x\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_MIPMAP_TEXTURE_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 1 || broken(value > 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_MIPMAP_LEVEL_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 0 || broken(value > 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_CUBE_MAP_FACE_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == WGL_TEXTURE_CUBE_MAP_POSITIVE_X_ARB || broken(value == 0xdeadbeef) /* AMD */, "got %#x\n", value );
-
-    pbuffer_attribs[0] = WGL_PBUFFER_WIDTH_ARB;
-    pbuffer_attribs[1] = 50;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_PBUFFER_HEIGHT_ARB;
-    pbuffer_attribs[1] = 50;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_PBUFFER_LOST_ARB;
-    pbuffer_attribs[1] = 0;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_TEXTURE_FORMAT_ARB;
-    pbuffer_attribs[1] = WGL_TEXTURE_RGBA_ARB;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_TEXTURE_TARGET_ARB;
-    pbuffer_attribs[1] = WGL_TEXTURE_2D_ARB;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_MIPMAP_TEXTURE_ARB;
-    pbuffer_attribs[1] = 2;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_MIPMAP_LEVEL_ARB;
-    pbuffer_attribs[1] = 2;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0 || broken(ret == 1) /* AMD */, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    pbuffer_attribs[0] = WGL_CUBE_MAP_FACE_ARB;
-    pbuffer_attribs[1] = WGL_TEXTURE_CUBE_MAP_NEGATIVE_X_ARB;
-    SetLastError( 0xdeadbeef );
-    ret = pwglSetPbufferAttribARB( pbuffer, pbuffer_attribs );
-    ok( ret == 0 || broken(ret == 1) /* AMD */, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_WIDTH_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 512 || broken(value == 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_HEIGHT_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 512 || broken(value == 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_PBUFFER_LOST_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 0, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_TEXTURE_FORMAT_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == WGL_TEXTURE_RGB_ARB || broken(value == 0xdeadbeef) /* AMD */, "got %#x\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_TEXTURE_TARGET_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == WGL_TEXTURE_CUBE_MAP_ARB || broken(value == 0xdeadbeef) /* AMD */, "got %#x\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_MIPMAP_TEXTURE_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    ok( value == 1 || broken(value > 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_MIPMAP_LEVEL_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    todo_wine ok( value == 0 || broken(value > 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    ret = pwglQueryPbufferARB( pbuffer, WGL_CUBE_MAP_FACE_ARB, &value );
-    ok( ret == 1, "got %u\n", ret );
-    todo_wine ok( value == WGL_TEXTURE_CUBE_MAP_POSITIVE_X_ARB || broken(value == 0xdeadbeef) /* AMD */, "got %#x\n", value );
-
-    pwglDestroyPbufferARB( pbuffer );
-
-
-    pbuffer_attribs[0] = WGL_TEXTURE_FORMAT_ARB;
-    pbuffer_attribs[1] = WGL_TEXTURE_RGB_ARB;
-    pbuffer_attribs[2] = WGL_TEXTURE_TARGET_ARB;
-    pbuffer_attribs[3] = WGL_TEXTURE_2D_ARB;
-    pbuffer_attribs[4] = 0;
-    pbuffer = pwglCreatePbufferARB( hdc, formats[0], 16, 16, pbuffer_attribs );
-    ok( !!pbuffer, "wglCreatePbufferARB returned %p\n", pbuffer );
-
-    pbuffer_dc = pwglGetPbufferDCARB( pbuffer );
-    ok( !!pbuffer_dc, "got %p\n", pbuffer_dc );
-    rc = wglCreateContext( pbuffer_dc );
-    ok( !!rc, "got %p\n", rc );
-    ret = wglMakeCurrent( pbuffer_dc, rc );
-    ok( ret == 1, "got %u\n", ret );
-
-    if (!winetest_platform_is_wine) /* triggers a BadMatch */
-    {
-    glClearColor( (float)0x22 / 0xff, (float)0x33 / 0xff, (float)0x44 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    }
-
-    ret = wglMakeCurrent( 0, 0 );
-    ok( ret == 1, "got %u\n", ret );
-    ret = wglDeleteContext( rc );
-    ok( ret == 1, "got %u\n", ret );
-    ret = pwglReleasePbufferDCARB( pbuffer, pbuffer_dc );
-    ok( ret == 1, "got %u\n", ret );
-
-
-    rc = wglCreateContext( hdc );
-    ok( !!rc, "got %p\n", rc );
-    ret = wglMakeCurrent( hdc, rc );
-    ok( ret == 1, "got %u\n", ret );
-
-    /* test some invalid params */
-    SetLastError( 0xdeadbeef );
-    ret = pwglReleaseTexImageARB( pbuffer, GL_FRONT );
-    todo_wine ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(GetLastError() == 0xdeadbeef) /* AMD */, "got %#lx\n", GetLastError() );
-    SetLastError( 0xdeadbeef );
-    ret = pwglBindTexImageARB( pbuffer, GL_BACK );
-    ok( ret == 0, "got %u\n", ret );
-    ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(GetLastError() == 0xdeadbeef) /* AMD */, "got %#lx\n", GetLastError() );
-
-    /* test invalid calls */
-    SetLastError( 0xdeadbeef );
-    ret = pwglReleaseTexImageARB( pbuffer, WGL_BACK_LEFT_ARB );
-    todo_wine ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-
-    value = 0xdeadbeef;
-    glGetIntegerv( GL_TEXTURE_BINDING_2D, &value );
-    ok( value == 0, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_WIDTH, &value );
-    ok( value == 0, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &value );
-    ok( value == 0, "got %u\n", value );
-    memset( pixels, 0xcd, sizeof(pixels) );
-    glGetTexImage( GL_TEXTURE_2D, 0, GL_RGB, GL_UNSIGNED_BYTE, pixels );
-    ok( pixels[0] == 0xcdcdcdcd, "got %#x\n", pixels[0] );
-    ret = pwglReleaseTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-
-    ret = pwglBindTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-
-    value = 0xdeadbeef;
-    glGetIntegerv( GL_TEXTURE_BINDING_2D, &value );
-    ok( value == 0, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_WIDTH, &value );
-    ok( value == 16 || broken(value == 0) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &value );
-    ok( value == 16 || broken(value == 0) /* AMD */, "got %u\n", value );
-    memset( pixels, 0xcd, sizeof(pixels) );
-    glGetTexImage( GL_TEXTURE_2D, 0, GL_RGB, GL_UNSIGNED_BYTE, pixels );
-    todo_wine ok( (pixels[0] & 0xffffff) == 0x443322 || broken(pixels[0] == 0xcdcdcdcd) /* AMD */, "got %#x\n", pixels[0] );
-
-    SetLastError( 0xdeadbeef );
-    ret = pwglBindTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    todo_wine ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-    SetLastError( 0xdeadbeef );
-    ret = pwglBindTexImageARB( pbuffer, WGL_FRONT_RIGHT_ARB );
-    todo_wine ok( ret == 0, "got %u\n", ret );
-    todo_wine ok( (GetLastError() & 0xffff) == ERROR_INVALID_DATA || broken(!GetLastError()) /* AMD */, "got %#lx\n", GetLastError() );
-
-    pwglReleaseTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    ret = pwglReleaseTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-
-    glGenTextures( 1, &texture );
-    glEnable( GL_TEXTURE_2D );
-    glBindTexture( GL_TEXTURE_2D, texture );
-    memset( pixels, 0xa5, sizeof(pixels) );
-    glTexImage2D( GL_TEXTURE_2D, 0, GL_RGBA8, 8, 8, 0,  GL_RGBA, GL_UNSIGNED_BYTE, pixels );
-
-    value = 0xdeadbeef;
-    glGetIntegerv( GL_TEXTURE_BINDING_2D, &value );
-    ok( value == texture, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_WIDTH, &value );
-    ok( value == 8, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &value );
-    ok( value == 8, "got %u\n", value );
-    memset( pixels, 0xcd, sizeof(pixels) );
-    glGetTexImage( GL_TEXTURE_2D, 0, GL_RGB, GL_UNSIGNED_BYTE, pixels );
-    ok( (pixels[0] & 0xffffff) == 0xa5a5a5, "got %#x\n", pixels[0] );
-
-    ret = pwglBindTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-
-    value = 0xdeadbeef;
-    glGetIntegerv( GL_TEXTURE_BINDING_2D, &value );
-    ok( value == texture, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_WIDTH, &value );
-    ok( value == 16 || broken(value == 8) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &value );
-    ok( value == 16 || broken(value == 8) /* AMD */, "got %u\n", value );
-    memset( pixels, 0xcd, sizeof(pixels) );
-    glGetTexImage( GL_TEXTURE_2D, 0, GL_RGB, GL_UNSIGNED_BYTE, pixels );
-    todo_wine ok( (pixels[0] & 0xffffff) == 0x443322 || broken(pixels[0] == 0xa5a5a5a5) /* AMD */, "got %#x\n", pixels[0] );
-
-    ret = pwglReleaseTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-
-    value = 0xdeadbeef;
-    glGetIntegerv( GL_TEXTURE_BINDING_2D, &value );
-    ok( value == texture, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_WIDTH, &value );
-    todo_wine ok( value == 0 || broken(value == 8) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &value );
-    todo_wine ok( value == 0 || broken(value == 8) /* AMD */, "got %u\n", value );
-    memset( pixels, 0xcd, sizeof(pixels) );
-    glGetTexImage( GL_TEXTURE_2D, 0, GL_RGB, GL_UNSIGNED_BYTE, pixels );
-    todo_wine ok( pixels[0] == 0xcdcdcdcd || broken(pixels[0] == 0xa5a5a5a5) /* AMD */, "got %#x\n", pixels[0] );
-
-    ret = pwglReleaseTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-    ret = pwglReleaseTexImageARB( pbuffer, WGL_FRONT_LEFT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-
-    ret = pwglBindTexImageARB( pbuffer, WGL_FRONT_RIGHT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-
-    value = 0xdeadbeef;
-    glGetIntegerv( GL_TEXTURE_BINDING_2D, &value );
-    ok( value == texture, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_WIDTH, &value );
-    todo_wine ok( value == 0 || broken(value == 8) /* AMD */, "got %u\n", value );
-    value = 0xdeadbeef;
-    glGetTexLevelParameteriv( GL_TEXTURE_2D, 0, GL_TEXTURE_HEIGHT, &value );
-    todo_wine ok( value == 0 || broken(value == 8) /* AMD */, "got %u\n", value );
-    memset( pixels, 0xcd, sizeof(pixels) );
-    glGetTexImage( GL_TEXTURE_2D, 0, GL_RGB, GL_UNSIGNED_BYTE, pixels );
-    todo_wine ok( pixels[0] == 0xcdcdcdcd || broken(pixels[0] == 0xa5a5a5a5) /* AMD */, "got %#x\n", pixels[0] );
-
-    ret = pwglReleaseTexImageARB( pbuffer, WGL_FRONT_RIGHT_ARB );
-    ok( ret == 1 || broken(ret == 0) /* AMD */, "got %u\n", ret );
-
-    glDeleteTextures( 1, &texture );
-
-    ret = wglDeleteContext( rc );
-    ok( ret == 1, "got %u\n", ret );
-
-    pwglDestroyPbufferARB( pbuffer );
-
-    ReleaseDC( hwnd, hdc );
-    DestroyWindow( hwnd );
-
-    wglMakeCurrent( old_hdc, old_rc );
+static void test_pbuffers(HDC hdc)
+{
+    const int iAttribList[] = { WGL_DRAW_TO_PBUFFER_ARB, 1, /* Request pbuffer support */
+                                0 };
+    int iFormats[MAX_FORMATS];
+    unsigned int nOnscreenFormats;
+    unsigned int nFormats;
+    int i, res;
+    int iPixelFormat = 0;
+
+    nOnscreenFormats = DescribePixelFormat(hdc, 0, 0, NULL);
+
+    /* When you want to render to a pbuffer you need to call wglGetPbufferDCARB which
+     * returns a 'magic' HDC which you can then pass to wglMakeCurrent to switch rendering
+     * to the pbuffer. Below some tests are performed on what happens if you use standard WGL calls
+     * on this 'magic' HDC for both a pixelformat that support onscreen and offscreen rendering
+     * and a pixelformat that's only available for offscreen rendering (this means that only
+     * wglChoosePixelFormatARB and friends know about the format.
+     *
+     * The first thing we need are pixelformats with pbuffer capabilities.
+     */
+    res = pwglChoosePixelFormatARB(hdc, iAttribList, NULL, MAX_FORMATS, iFormats, &nFormats);
+    if(res <= 0)
+    {
+        skip("No pbuffer compatible formats found while WGL_ARB_pbuffer is supported\n");
+        return;
+    }
+    trace("nOnscreenFormats: %d\n", nOnscreenFormats);
+    trace("Total number of pbuffer capable pixelformats: %d\n", nFormats);
+
+    /* Try to select an onscreen pixelformat out of the list */
+    for(i=0; i < nFormats; i++)
+    {
+        /* Check if the format is onscreen, if it is choose it */
+        if(iFormats[i] <= nOnscreenFormats)
+        {
+            iPixelFormat = iFormats[i];
+            trace("Selected iPixelFormat=%d\n", iPixelFormat);
+            break;
+        }
+    }
+
+    /* A video driver supports a large number of onscreen and offscreen pixelformats.
+     * The traditional WGL calls only see a subset of the whole pixelformat list. First
+     * of all they only see the onscreen formats (the offscreen formats are at the end of the
+     * pixelformat list) and second extended pixelformat capabilities are hidden from the
+     * standard WGL calls. Only functions that depend on WGL_ARB_pixel_format can see them.
+     *
+     * Below we check if the pixelformat is also supported onscreen.
+     */
+    if(iPixelFormat != 0)
+    {
+        HDC pbuffer_hdc;
+        int attrib = 0;
+        HPBUFFERARB pbuffer = pwglCreatePbufferARB(hdc, iPixelFormat, 640 /* width */, 480 /* height */, &attrib);
+        if(!pbuffer)
+            skip("Pbuffer creation failed!\n");
+
+        /* Test the pixelformat returned by GetPixelFormat on a pbuffer as the behavior is not clear */
+        pbuffer_hdc = pwglGetPbufferDCARB(pbuffer);
+        res = GetPixelFormat(pbuffer_hdc);
+        ok(res == iPixelFormat, "Unexpected iPixelFormat=%d returned by GetPixelFormat for format %d\n", res, iPixelFormat);
+        trace("iPixelFormat returned by GetPixelFormat: %d\n", res);
+        trace("PixelFormat from wglChoosePixelFormatARB: %d\n", iPixelFormat);
+
+        pwglReleasePbufferDCARB(pbuffer, pbuffer_hdc);
+    }
+    else skip("Pbuffer test for onscreen pixelformat skipped as no onscreen format with pbuffer capabilities have been found\n");
+
+    /* Search for a real offscreen format */
+    for(i=0, iPixelFormat=0; i<nFormats; i++)
+    {
+        if(iFormats[i] > nOnscreenFormats)
+        {
+            iPixelFormat = iFormats[i];
+            trace("Selected iPixelFormat: %d\n", iPixelFormat);
+            break;
+        }
+    }
+
+    if(iPixelFormat != 0)
+    {
+        HDC pbuffer_hdc;
+        HPBUFFERARB pbuffer = pwglCreatePbufferARB(hdc, iPixelFormat, 640 /* width */, 480 /* height */, NULL);
+        if(pbuffer)
+        {
+            /* Test the pixelformat returned by GetPixelFormat on a pbuffer as the behavior is not clear */
+            pbuffer_hdc = pwglGetPbufferDCARB(pbuffer);
+            res = GetPixelFormat(pbuffer_hdc);
+
+            ok(res == 1, "Unexpected iPixelFormat=%d (1 expected) returned by GetPixelFormat for offscreen format %d\n", res, iPixelFormat);
+            trace("iPixelFormat returned by GetPixelFormat: %d\n", res);
+            trace("PixelFormat from wglChoosePixelFormatARB: %d\n", iPixelFormat);
+            pwglReleasePbufferDCARB(pbuffer, hdc);
+        }
+        else skip("Pbuffer creation failed!\n");
+    }
+    else skip("Pbuffer test for offscreen pixelformat skipped as no offscreen-only format with pbuffer capabilities has been found\n");
 }
 
 static int test_pfd(const PIXELFORMATDESCRIPTOR *pfd, PIXELFORMATDESCRIPTOR *fmt)
@@ -1193,24 +647,8 @@
 
                     if (source_current)
                     {
-                        float floats[4] = {1.0f,0.0f,1.0f,0.0f};
-
                         res = wglMakeCurrent(winhdc, source);
                         ok(res, "Make source current failed\n");
-
-                        glViewport(0, 0, 256, 256);
-                        glLightModelfv(GL_LIGHT_MODEL_AMBIENT, floats);
-                        glLightModeli(GL_LIGHT_MODEL_TWO_SIDE, 0);
-                        glEnable(GL_NORMALIZE);
-                        glEnable(GL_DEPTH_TEST);
-                        glEnable(GL_CULL_FACE);
-                        glEnable(GL_LIGHTING);
-                        glDisable(GL_FOG);
-                        glDisable(GL_DITHER);
-                        glDepthFunc(GL_LESS);
-                        glHint(GL_PERSPECTIVE_CORRECTION_HINT, GL_NICEST);
-                        glShadeModel(GL_SMOOTH);
-                        glClearColor(0.1, 0.2, 0.3, 1.0);
                     }
                     if (source_sharing)
                     {
@@ -1219,119 +657,20 @@
                     }
                     if (dest_current)
                     {
-                        float floats[4] = {0.0f,1.0f,0.0f,1.0f};
-
                         res = wglMakeCurrent(winhdc, dest);
                         ok(res, "Make dest current failed\n");
-
-                        glViewport(0, 0, 128, 128);
-                        glLightModelfv(GL_LIGHT_MODEL_AMBIENT, floats);
-                        glLightModeli(GL_LIGHT_MODEL_TWO_SIDE, 1);
-                        glDisable(GL_NORMALIZE);
-                        glDisable(GL_DEPTH_TEST);
-                        glDisable(GL_CULL_FACE);
-                        glDisable(GL_LIGHTING);
-                        glEnable(GL_FOG);
-                        glEnable(GL_DITHER);
-                        glDepthFunc(GL_GREATER);
-                        glHint(GL_PERSPECTIVE_CORRECTION_HINT, GL_NICEST);
-                        glShadeModel(GL_FLAT);
-                        glClearColor(0.3, 0.2, 0.1, 1.0);
                     }
                     if (dest_sharing)
                     {
                         res = wglShareLists(other, dest);
+                        todo_wine_if(source_sharing && dest_current)
                         ok(res, "Sharing of display lists from other to dest failed\n");
                     }
 
                     res = wglShareLists(source, dest);
+                    todo_wine_if((source_current || source_sharing) && (dest_current || dest_sharing))
                     ok(res || broken(nvidia && !source_sharing && dest_sharing),
                        "Sharing of display lists from source to dest failed\n");
-
-                    if (source_current)
-                    {
-                        float floats[4];
-                        int ints[4];
-
-                        res = wglMakeCurrent(winhdc, source);
-                        ok(res, "Make source current failed\n");
-
-                        glGetIntegerv(GL_VIEWPORT, ints);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        ok(ints[1] == 0, "got %d\n", ints[1]);
-                        ok(ints[2] == 256, "got %d\n", ints[2]);
-                        ok(ints[3] == 256, "got %d\n", ints[3]);
-                        glGetFloatv(GL_LIGHT_MODEL_AMBIENT, floats);
-                        ok(floats[0] == 1.0f, "got %f\n", floats[0]);
-                        ok(floats[1] == 0.0f, "got %f\n", floats[1]);
-                        ok(floats[2] == 1.0f, "got %f\n", floats[2]);
-                        ok(floats[3] == 0.0f, "got %f\n", floats[3]);
-                        glGetIntegerv(GL_LIGHT_MODEL_TWO_SIDE, ints);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_NORMALIZE);
-                        ok(ints[0] == 1, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_DEPTH_TEST);
-                        ok(ints[0] == 1, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_CULL_FACE);
-                        ok(ints[0] == 1, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_LIGHTING);
-                        ok(ints[0] == 1, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_FOG);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_DITHER);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        glGetIntegerv(GL_DEPTH_FUNC, ints);
-                        ok(ints[0] == GL_LESS, "got %d\n", ints[0]);
-                        glGetIntegerv(GL_SHADE_MODEL, ints);
-                        ok(ints[0] == GL_SMOOTH, "got %d\n", ints[0]);
-                        glGetFloatv(GL_COLOR_CLEAR_VALUE, floats);
-                        ok(floats[0] == 0.1f, "got %f\n", floats[0]);
-                        ok(floats[1] == 0.2f, "got %f\n", floats[1]);
-                        ok(floats[2] == 0.3f, "got %f\n", floats[2]);
-                        ok(floats[3] == 1.0f, "got %f\n", floats[3]);
-                    }
-                    if (dest_current)
-                    {
-                        float floats[4];
-                        int ints[4];
-
-                        res = wglMakeCurrent(winhdc, dest);
-                        ok(res, "Make dest current failed\n");
-
-                        glGetIntegerv(GL_VIEWPORT, ints);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        ok(ints[1] == 0, "got %d\n", ints[1]);
-                        ok(ints[2] == 128, "got %d\n", ints[2]);
-                        ok(ints[3] == 128, "got %d\n", ints[3]);
-                        glGetFloatv(GL_LIGHT_MODEL_AMBIENT, floats);
-                        ok(floats[0] == 0.0f, "got %f\n", floats[0]);
-                        ok(floats[1] == 1.0f, "got %f\n", floats[1]);
-                        ok(floats[2] == 0.0f, "got %f\n", floats[2]);
-                        ok(floats[3] == 1.0f, "got %f\n", floats[3]);
-                        glGetIntegerv(GL_LIGHT_MODEL_TWO_SIDE, ints);
-                        ok(ints[0] == 1, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_NORMALIZE);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_DEPTH_TEST);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_CULL_FACE);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_LIGHTING);
-                        ok(ints[0] == 0, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_FOG);
-                        ok(ints[0] == 1, "got %d\n", ints[0]);
-                        ints[0] = glIsEnabled(GL_DITHER);
-                        ok(ints[0] == 1, "got %d\n", ints[0]);
-                        glGetIntegerv(GL_DEPTH_FUNC, ints);
-                        ok(ints[0] == GL_GREATER, "got %d\n", ints[0]);
-                        glGetIntegerv(GL_SHADE_MODEL, ints);
-                        ok(ints[0] == GL_FLAT, "got %d\n", ints[0]);
-                        glGetFloatv(GL_COLOR_CLEAR_VALUE, floats);
-                        ok(floats[0] == 0.3f, "got %f\n", floats[0]);
-                        ok(floats[1] == 0.2f, "got %f\n", floats[1]);
-                        ok(floats[2] == 0.1f, "got %f\n", floats[2]);
-                        ok(floats[3] == 1.0f, "got %f\n", floats[3]);
-                    }
 
                     if (source_current || dest_current)
                     {
@@ -1520,587 +859,155 @@
     }
 }
 
-static void read_bitmap_pixels( HDC hdc, HBITMAP bmp, UINT *pixels, UINT width, UINT height, UINT bpp )
-{
-    BITMAPINFO bmi =
-    {
-        .bmiHeader.biSize = sizeof(BITMAPINFOHEADER),
-        .bmiHeader.biPlanes = 1,
-        .bmiHeader.biWidth = width,
-        .bmiHeader.biHeight = -height,
-        .bmiHeader.biBitCount = bpp,
-        .bmiHeader.biSizeImage = width * height * bpp / 8,
-        .bmiHeader.biCompression = BI_RGB,
-    };
-    BOOL ret;
-
-    ret = GetDIBits( hdc, bmp, 0, height, pixels, &bmi, DIB_RGB_COLORS );
-    ok( ret, "GetDIBits failed, error %lu\n", GetLastError() );
-}
-
 static void test_bitmap_rendering( BOOL use_dib )
 {
-    static const RECT expect_rect = {0, 0, 4, 4}, expect_rect2 = {0, 0, 12, 12};
-    BITMAPINFO bmi = {.bmiHeader = {.biSize = sizeof(BITMAPINFOHEADER), .biPlanes = 1, .biCompression = BI_RGB}};
-    UINT buffer[16 * 16], buffer2[16 * 16], *pixels = buffer, *pixels2 = buffer2, pixel;
-    int i, ret, bpp, count, pixel_format = 0;
-    HBITMAP bmp, old_bmp, bmp2, tmp_bmp;
-    GLint viewport[4], object;
+    PIXELFORMATDESCRIPTOR pfd;
+    int i, ret, bpp, iPixelFormat=0;
+    unsigned int nFormats;
     HGLRC hglrc, hglrc2;
-    HDC hdc;
-
-    winetest_push_context( use_dib ? "DIB" : "DDB" );
-
-    hdc = CreateCompatibleDC( 0 );
+    BITMAPINFO biDst;
+    HBITMAP bmpDst, oldDst, bmp2;
+    HDC hdcDst, hdcScreen;
+    UINT *dstBuffer = NULL;
+
+    hdcScreen = CreateCompatibleDC(0);
+    hdcDst = CreateCompatibleDC(0);
 
     if (use_dib)
     {
         bpp = 32;
-        bmi.bmiHeader.biWidth = 4;
-        bmi.bmiHeader.biHeight = -4;
-        bmi.bmiHeader.biBitCount = 32;
-        bmp = CreateDIBSection( 0, &bmi, DIB_RGB_COLORS, (void **)&pixels, NULL, 0 );
-        memset( (void *)pixels, 0xcd, sizeof(*pixels) * 4 * 4 );
-
-        bmi.bmiHeader.biWidth = 12;
-        bmi.bmiHeader.biHeight = -12;
-        bmi.bmiHeader.biBitCount = 16;
-        bmp2 = CreateDIBSection( 0, &bmi, DIB_RGB_COLORS, (void **)&pixels2, NULL, 0 );
-        memset( (void *)pixels2, 0xdc, sizeof(*pixels2) * 12 * 12 );
+        memset(&biDst, 0, sizeof(BITMAPINFO));
+        biDst.bmiHeader.biSize = sizeof(BITMAPINFOHEADER);
+        biDst.bmiHeader.biWidth = 4;
+        biDst.bmiHeader.biHeight = -4;
+        biDst.bmiHeader.biPlanes = 1;
+        biDst.bmiHeader.biBitCount = 32;
+        biDst.bmiHeader.biCompression = BI_RGB;
+
+        bmpDst = CreateDIBSection(0, &biDst, DIB_RGB_COLORS, (void**)&dstBuffer, NULL, 0);
+
+        biDst.bmiHeader.biWidth = 12;
+        biDst.bmiHeader.biHeight = -12;
+        biDst.bmiHeader.biBitCount = 16;
+        bmp2 = CreateDIBSection(0, &biDst, DIB_RGB_COLORS, NULL, NULL, 0);
     }
     else
     {
-        bpp = GetDeviceCaps( hdc, BITSPIXEL );
-        memset( (void *)pixels, 0xcd, sizeof(*pixels) * 4 * 4 );
-        bmp = CreateBitmap( 4, 4, 1, bpp, pixels );
-        memset( (void *)pixels2, 0xdc, sizeof(*pixels2) * 12 * 12 );
-        bmp2 = CreateBitmap( 12, 12, 1, bpp, pixels2 );
-    }
-
-    ret = GetPixelFormat( hdc );
-    ok( ret == 0, "got %d\n", ret );
-    count = DescribePixelFormat( hdc, 0, 0, NULL );
-    ok( count > 1, "got %d\n", count );
-
-    old_bmp = SelectObject( hdc, bmp );
-    ok( !!old_bmp, "got %p\n", old_bmp );
-
-    /* cannot create a GL context without a pixel format */
-
-    hglrc = wglCreateContext( hdc );
-    ok( !hglrc, "wglCreateContext succeeded\n" );
-    if (hglrc) wglDeleteContext( hglrc );
-
-    /* cannot set pixel format twice */
-
-    for (i = 1; i <= count; i++)
-    {
-        PIXELFORMATDESCRIPTOR pfd = {0};
-
-        winetest_push_context( "%u", i );
-
-        ret = DescribePixelFormat( hdc, i, sizeof(pfd), &pfd );
-        ok( ret == count, "got %d\n", ret );
-
-        if ((pfd.dwFlags & PFD_DRAW_TO_BITMAP) && (pfd.dwFlags & PFD_SUPPORT_OPENGL) &&
-            pfd.cColorBits == bpp && pfd.cAlphaBits > 0)
+        bpp = GetDeviceCaps( hdcScreen, BITSPIXEL );
+        bmpDst = CreateBitmap( 4, 4, 1, bpp, NULL );
+        bmp2 = CreateBitmap( 12, 12, 1, bpp, NULL );
+    }
+
+    oldDst = SelectObject(hdcDst, bmpDst);
+
+    trace( "testing on %s\n", use_dib ? "DIB" : "DDB" );
+
+    /* Pick a pixel format by hand because ChoosePixelFormat is unreliable */
+    nFormats = DescribePixelFormat(hdcDst, 0, 0, NULL);
+    for(i=1; i<=nFormats; i++)
+    {
+        memset(&pfd, 0, sizeof(PIXELFORMATDESCRIPTOR));
+        DescribePixelFormat(hdcDst, i, sizeof(PIXELFORMATDESCRIPTOR), &pfd);
+
+        if((pfd.dwFlags & PFD_DRAW_TO_BITMAP) &&
+           (pfd.dwFlags & PFD_SUPPORT_OPENGL) &&
+           (pfd.cColorBits == bpp) &&
+           (pfd.cAlphaBits == 8) )
         {
-            ret = SetPixelFormat( hdc, i, &pfd );
-            if (pixel_format) ok( !ret, "SetPixelFormat succeeded\n" );
-            else ok( ret, "SetPixelFormat failed\n" );
-            if (ret) pixel_format = i;
-            ret = GetPixelFormat( hdc );
-            ok( ret == pixel_format, "got %d\n", ret );
+            iPixelFormat = i;
+            break;
         }
-
-        winetest_pop_context();
-    }
-
-    ok( !!pixel_format, "got pixel_format %u\n", pixel_format );
-
-    /* even after changing the selected bitmap */
-
-    tmp_bmp = SelectObject( hdc, bmp2 );
-    ok( tmp_bmp == bmp, "got %p\n", tmp_bmp );
-
-    for (i = 1; i <= count; i++)
-    {
-        PIXELFORMATDESCRIPTOR pfd = {0};
-
-        winetest_push_context( "%u", i );
-
-        ret = DescribePixelFormat( hdc, i, sizeof(pfd), &pfd );
-        ok( ret == count, "got %d\n", ret );
-
-        ret = SetPixelFormat( hdc, i, &pfd );
-        if (pixel_format != i) ok( !ret, "SetPixelFormat succeeded\n" );
-        else ok( ret, "SetPixelFormat succeeded\n" );
-        ret = GetPixelFormat( hdc );
-        ok( ret == pixel_format, "got %d\n", ret );
-
-        winetest_pop_context();
-    }
-
-    tmp_bmp = SelectObject( hdc, bmp );
-    ok( tmp_bmp == bmp2, "got %p\n", tmp_bmp );
-
-    /* creating a GL context now works */
-
-    hglrc = wglCreateContext( hdc );
-    ok( !!hglrc, "wglCreateContext failed, error %lu\n", GetLastError() );
-    ret = wglMakeCurrent( hdc, hglrc );
-    ok( ret, "wglMakeCurrent failed, error %lu\n", GetLastError() );
-
-    glGetIntegerv( GL_READ_BUFFER, &object );
-    ok( object == GL_FRONT, "got %u\n", object );
-    glGetIntegerv( GL_DRAW_BUFFER, &object );
-    ok( object == GL_FRONT, "got %u\n", object );
-
-    memset( viewport, 0xcd, sizeof(viewport) );
-    glGetIntegerv( GL_VIEWPORT, viewport );
-    ok( EqualRect( (RECT *)viewport, &expect_rect ), "got viewport %s\n", wine_dbgstr_rect( (RECT *)viewport ) );
-
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0xcdcdcd, "got %#x\n", pixel );
-
-    glClearColor( (float)0x22 / 0xff, (float)0x33 / 0xff, (float)0x44 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0xcdcdcd, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0xdcdcdc, "got %#x\n", pixels2[0] );
-
-    glFinish();
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x223344, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0xdcdcdc, "got %#x\n", pixels2[0] );
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0x443322, "got %#x\n", pixel );
-
-
-    glClearColor( (float)0x55 / 0xff, (float)0x66 / 0xff, (float)0x77 / 0xff, (float)0x88 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x223344, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0xdcdcdc, "got %#x\n", pixels2[0] );
-
-    glFlush();
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x556677, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0xdcdcdc, "got %#x\n", pixels2[0] );
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0x776655, "got %#x\n", pixel );
-
-
-    glClearColor( (float)0x22 / 0xff, (float)0x33 / 0xff, (float)0x44 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x556677, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0xdcdcdc, "got %#x\n", pixels2[0] );
-
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0x443322, "got %#x\n", pixel );
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x223344, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0xdcdcdc, "got %#x\n", pixels2[0] );
-
-
-    tmp_bmp = SelectObject( hdc, bmp2 );
-    ok( tmp_bmp == bmp, "got %p\n", tmp_bmp );
-
-    /* context still uses the old pixel format and viewport */
-    memset( viewport, 0xcd, sizeof(viewport) );
-    glGetIntegerv( GL_VIEWPORT, viewport );
-    ok( EqualRect( (RECT *)viewport, &expect_rect ), "got viewport %s\n", wine_dbgstr_rect( (RECT *)viewport ) );
-
-    /* pixels are read from the selected bitmap */
-
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0xdcdcdc, "got %#x\n", pixel );
-
-    if (use_dib)
-    {
-        memset( buffer2, 0xa5, sizeof(buffer2) );
-        glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-        ok( (pixel & 0xffffff) == 0xdcdcdc, "got %#x\n", pixel );
-        memset( buffer2, 0xdc, sizeof(buffer2) );
-    }
-
-    /* GL doesn't render to the bitmap that was selected on wglMakeCurrent, but
-     * copies to the bitmap that is currently selected on the HDC on Finish/Flush.
-     */
-
-    glClearColor( (float)0x44 / 0xff, (float)0x33 / 0xff, (float)0x22 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x223344, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0xdcdcdc, "got %#x\n", pixels2[0] );
-
-    glFinish();
-
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0x223344, "got %#x\n", pixel );
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x223344, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0x443322, "got %#x\n", pixels2[0] );
-
-
-    ret = wglMakeCurrent( NULL, NULL );
-    ok( ret, "wglMakeCurrent failed, error %lu\n", GetLastError() );
-    ret = wglMakeCurrent( hdc, hglrc );
-    ok( ret, "wglMakeCurrent failed, error %lu\n", GetLastError() );
-
-    memset( viewport, 0xcd, sizeof(viewport) );
-    glGetIntegerv( GL_VIEWPORT, viewport );
-    ok( EqualRect( (RECT *)viewport, &expect_rect ), "got viewport %s\n", wine_dbgstr_rect( (RECT *)viewport ) );
-
-    glClearColor( (float)0x44 / 0xff, (float)0x55 / 0xff, (float)0x66 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    glFinish();
-
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x223344, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0x445566, "got %#x\n", pixels2[0] );
-
-
-    /* creating a context uses the currently selected bitmap size as viewport */
-
-    hglrc2 = wglCreateContext( hdc );
-    ok( !!hglrc2, "wglCreateContext failed, error %lu\n", GetLastError() );
-
-    ret = wglMakeCurrent( hdc, hglrc2 );
-    ok( ret, "wglMakeCurrent failed, error %lu\n", GetLastError() );
-
-    memset( viewport, 0xcd, sizeof(viewport) );
-    glGetIntegerv( GL_VIEWPORT, viewport );
-    ok( EqualRect( (RECT *)viewport, &expect_rect2 ), "got viewport %s\n", wine_dbgstr_rect( (RECT *)viewport ) );
-
-    glClearColor( (float)0x66 / 0xff, (float)0x55 / 0xff, (float)0x44 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    glFinish();
-
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x223344, "got %#x\n", pixels[0] );
-    if (use_dib) todo_wine ok( (pixels2[0] & 0xffffff) == 0x03148, "got %#x\n", pixels2[0] );
-    else ok( (pixels2[0] & 0xffffff) == 0x665544, "got %#x\n", pixels2[0] );
-
-    ret = wglMakeCurrent( hdc, hglrc );
-    ok( ret, "wglMakeCurrent failed, error %lu\n", GetLastError() );
-
-    memset( viewport, 0xcd, sizeof(viewport) );
-    glGetIntegerv( GL_VIEWPORT, viewport );
-    ok( EqualRect( (RECT *)viewport, &expect_rect ), "got viewport %s\n", wine_dbgstr_rect( (RECT *)viewport ) );
-
-    glClearColor( (float)0x66 / 0xff, (float)0x77 / 0xff, (float)0x88 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    glFinish();
-
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    ok( (pixels[0] & 0xffffff) == 0x223344, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0x667788, "got %#x\n", pixels2[0] );
-
-
-    tmp_bmp = SelectObject( hdc, bmp );
-    ok( tmp_bmp == bmp2, "got %p\n", tmp_bmp );
-
-    ret = wglMakeCurrent( hdc, hglrc2 );
-    ok( ret, "wglMakeCurrent failed, error %lu\n", GetLastError() );
-
-    memset( viewport, 0xcd, sizeof(viewport) );
-    glGetIntegerv( GL_VIEWPORT, viewport );
-    ok( EqualRect( (RECT *)viewport, &expect_rect2 ), "got viewport %s\n", wine_dbgstr_rect( (RECT *)viewport ) );
-
-    glClearColor( (float)0x88 / 0xff, (float)0x77 / 0xff, (float)0x66 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    glFinish();
-
-    if (pixels == buffer) read_bitmap_pixels( hdc, bmp, pixels, 4, 4, bpp );
-    if (pixels2 == buffer2) read_bitmap_pixels( hdc, bmp2, pixels2, 12, 12, bpp );
-    if (use_dib) todo_wine ok( (pixels[0] & 0xffffff) == 0x45cc, "got %#x\n", pixels[0] );
-    else ok( (pixels[0] & 0xffffff) == 0x887766, "got %#x\n", pixels[0] );
-    ok( (pixels2[0] & 0xffffff) == 0x667788, "got %#x\n", pixels2[0] );
-
-    wglDeleteContext( hglrc2 );
-    wglDeleteContext( hglrc );
-
-    SelectObject( hdc, old_bmp );
-    DeleteObject( bmp2 );
-    DeleteObject( bmp );
-    DeleteDC( hdc );
-
-    winetest_pop_context();
-}
-
-static void test_16bit_bitmap_rendering(void)
-{
-    PIXELFORMATDESCRIPTOR pfd;
-    INT pixel_format, success;
-    HGDIOBJ old_gdi_obj;
-    USHORT *pixels;
-    HBITMAP bitmap;
-    HGLRC gl;
-    HDC hdc;
-
-    PIXELFORMATDESCRIPTOR pixel_format_args = {
-        .nSize = sizeof(PIXELFORMATDESCRIPTOR),
-        .nVersion = 1,
-        .dwFlags = PFD_DRAW_TO_BITMAP | PFD_SUPPORT_OPENGL | PFD_DEPTH_DONTCARE,
-        .iPixelType = PFD_TYPE_RGBA,
-        .iLayerType = PFD_MAIN_PLANE,
-        .cColorBits = 16,
-        .cAlphaBits = 0
-    };
-    BITMAPINFO bitmap_args = {
-        .bmiHeader = {
-            .biSize = sizeof(BITMAPINFOHEADER),
-            .biPlanes = 1,
-            .biCompression = BI_RGB,
-            .biWidth = 4,
-            .biHeight = -4,  /* Four pixels tall with the origin in the top-left corner. */
-            .biBitCount = 16
+    }
+
+    if(!iPixelFormat)
+    {
+        skip("Unable to find a suitable pixel format\n");
+    }
+    else
+    {
+        ret = SetPixelFormat(hdcDst, iPixelFormat, &pfd);
+        ok( ret, "SetPixelFormat failed\n" );
+        ret = GetPixelFormat( hdcDst );
+        ok( ret == iPixelFormat, "GetPixelFormat returned %d/%d\n", ret, iPixelFormat );
+        ret = SetPixelFormat(hdcDst, iPixelFormat + 1, &pfd);
+        ok( !ret, "SetPixelFormat succeeded\n" );
+        hglrc = wglCreateContext(hdcDst);
+        ok(hglrc != NULL, "Unable to create a context\n");
+
+        if(hglrc)
+        {
+            GLint viewport[4];
+            wglMakeCurrent(hdcDst, hglrc);
+            hglrc2 = wglCreateContext(hdcDst);
+            ok(hglrc2 != NULL, "Unable to create a context\n");
+
+            /* Note this is RGBA but we read ARGB back */
+            glClearColor((float)0x22/0xff, (float)0x33/0xff, (float)0x44/0xff, (float)0x11/0xff);
+            glClear(GL_COLOR_BUFFER_BIT);
+            glGetIntegerv( GL_VIEWPORT, viewport );
+            glFinish();
+
+            ok( viewport[0] == 0 && viewport[1] == 0 && viewport[2] == 4 && viewport[3] == 4,
+                "wrong viewport %d,%d,%d,%d\n", viewport[0], viewport[1], viewport[2], viewport[3] );
+            /* Note apparently the alpha channel is not supported by the software renderer (bitmap only works using software) */
+            if (dstBuffer)
+                for (i = 0; i < 16; i++)
+                    ok(dstBuffer[i] == 0x223344 || dstBuffer[i] == 0x11223344, "Received color=%x at %u\n",
+                       dstBuffer[i], i);
+
+            SelectObject(hdcDst, bmp2);
+            ret = GetPixelFormat( hdcDst );
+            ok( ret == iPixelFormat, "GetPixelFormat returned %d/%d\n", ret, iPixelFormat );
+            ret = SetPixelFormat(hdcDst, iPixelFormat + 1, &pfd);
+            ok( !ret, "SetPixelFormat succeeded\n" );
+
+            /* context still uses the old pixel format and viewport */
+            glClearColor((float)0x44/0xff, (float)0x33/0xff, (float)0x22/0xff, (float)0x11/0xff);
+            glClear(GL_COLOR_BUFFER_BIT);
+            glFinish();
+            glGetIntegerv( GL_VIEWPORT, viewport );
+            ok( viewport[0] == 0 && viewport[1] == 0 && viewport[2] == 4 && viewport[3] == 4,
+                "wrong viewport %d,%d,%d,%d\n", viewport[0], viewport[1], viewport[2], viewport[3] );
+
+            wglMakeCurrent(NULL, NULL);
+            wglMakeCurrent(hdcDst, hglrc);
+            glClearColor((float)0x44/0xff, (float)0x55/0xff, (float)0x66/0xff, (float)0x11/0xff);
+            glClear(GL_COLOR_BUFFER_BIT);
+            glFinish();
+            glGetIntegerv( GL_VIEWPORT, viewport );
+            ok( viewport[0] == 0 && viewport[1] == 0 && viewport[2] == 4 && viewport[3] == 4,
+                "wrong viewport %d,%d,%d,%d\n", viewport[0], viewport[1], viewport[2], viewport[3] );
+
+            wglMakeCurrent(hdcDst, hglrc2);
+            glGetIntegerv( GL_VIEWPORT, viewport );
+            ok( viewport[0] == 0 && viewport[1] == 0 && viewport[2] == 12 && viewport[3] == 12,
+                "wrong viewport %d,%d,%d,%d\n", viewport[0], viewport[1], viewport[2], viewport[3] );
+
+            wglMakeCurrent(hdcDst, hglrc);
+            glGetIntegerv( GL_VIEWPORT, viewport );
+            ok( viewport[0] == 0 && viewport[1] == 0 && viewport[2] == 4 && viewport[3] == 4,
+                "wrong viewport %d,%d,%d,%d\n", viewport[0], viewport[1], viewport[2], viewport[3] );
+
+            SelectObject(hdcDst, bmpDst);
+            ret = GetPixelFormat( hdcDst );
+            ok( ret == iPixelFormat, "GetPixelFormat returned %d/%d\n", ret, iPixelFormat );
+            ret = SetPixelFormat(hdcDst, iPixelFormat + 1, &pfd);
+            ok( !ret, "SetPixelFormat succeeded\n" );
+            wglMakeCurrent(hdcDst, hglrc2);
+            glGetIntegerv( GL_VIEWPORT, viewport );
+            ok( viewport[0] == 0 && viewport[1] == 0 && viewport[2] == 12 && viewport[3] == 12,
+                "wrong viewport %d,%d,%d,%d\n", viewport[0], viewport[1], viewport[2], viewport[3] );
+
+            wglDeleteContext(hglrc2);
+            wglDeleteContext(hglrc);
         }
-    };
-
-    hdc = CreateCompatibleDC(NULL);
-    ok(hdc != NULL, "Failed to get a device context\n");
-
-    /* Create a bitmap. */
-    bitmap = CreateDIBSection(NULL, &bitmap_args, DIB_RGB_COLORS, (void**)&pixels, NULL, 0);
-    old_gdi_obj = SelectObject(hdc, bitmap);
-    ok(old_gdi_obj != NULL, "Failed to SetObject\n");
-
-    /* Choose a pixel format. */
-    pixel_format = ChoosePixelFormat(hdc, &pixel_format_args);
-    todo_wine ok(pixel_format != 0, "Failed to get a 16 bit pixel format with the DRAW_TO_BITMAP flag.\n");
-
-    if (pixel_format == 0)
-    {
-        skip("Skipping 16-bit rendering test"
-                " (no 16 bit pixel format with the DRAW_TO_BITMAP flag was available)\n");
-        SelectObject(hdc, old_gdi_obj);
-        DeleteObject(bitmap);
-        DeleteDC(hdc);
-        return;
-    }
-
-    /* When asking for a 16-bit DRAW_TO_BITMAP pixel format, Windows will give you r5g5b5a1 by
-     * default, even if you didn't ask for an alpha bit.
-     *
-     * It's important to note that all of the color bits have to match exactly, because the renders
-     * are sent back to the CPU and will have to match any other software rendering operations that
-     * the program does (DRAW_TO_BITMAP is normally used in combination with blitting). */
-    success = DescribePixelFormat(hdc, pixel_format, sizeof(pfd), &pfd);
-    ok(success != 0, "Failed to DescribePixelFormat (error: %lu)\n", GetLastError());
-    /* Likely MSDN inaccuracy: According to the PIXELFORMATDESCRIPTOR docs, alpha bits are excluded
-     * from cColorBits. It doesn't seem like that's true. */
-    ok(pfd.cColorBits == 16, "Wrong amount of color bits (got %d, expected 16)\n", pfd.cColorBits);
-    todo_wine ok(pfd.cRedBits == 5, "Wrong amount of red bits (got %d, expected 5)\n", pfd.cRedBits);
-    todo_wine ok(pfd.cGreenBits == 5, "Wrong amount of green bits (got %d, expected 5)\n", pfd.cGreenBits);
-    todo_wine ok(pfd.cBlueBits == 5, "Wrong amount of blue bits (got %d, expected 5)\n", pfd.cBlueBits);
-    /* Quirky: It seems that there's an alpha bit, but it somehow doesn't count as one for
-     * DescribePixelFormat. On Windows cAlphaBits is zero.
-     * ok(pfd.cAlphaBits == 1, "Wrong amount of alpha bits (got %d, expected 1)\n", pfd.cAlphaBits); */
-    todo_wine ok(pfd.cRedShift == 10, "Wrong red shift (got %d, expected 10)\n", pfd.cRedShift);
-    todo_wine ok(pfd.cGreenShift == 5, "Wrong green shift (got %d, expected 5)\n", pfd.cGreenShift);
-    /* This next test might fail, depending on your drivers. */
-    ok(pfd.cBlueShift == 0, "Wrong blue shift (got %d, expected 0)\n", pfd.cBlueShift);
-
-    success = SetPixelFormat(hdc, pixel_format, &pixel_format_args);
-    ok(success, "Failed to SetPixelFormat (error: %lu)\n", GetLastError());
-
-    /* Create an OpenGL context. */
-    gl = wglCreateContext(hdc);
-    ok(gl != NULL, "Failed to wglCreateContext (error: %lu)\n", GetLastError());
-    success = wglMakeCurrent(hdc, gl);
-    ok(success, "Failed to wglMakeCurrent (error: %lu)\n", GetLastError());
-
-    /* Try setting the bitmap to white. */
-    glClearColor(1.0f, 1.0f, 1.0f, 1.0f);
-    glClear(GL_COLOR_BUFFER_BIT);
-    glFinish();
-    todo_wine ok(pixels[0] == 0x7fff, "Wrong color after glClear at (0, 0): %#x\n", pixels[0]);
-    todo_wine ok(pixels[1] == 0x7fff, "Wrong color after glClear at (1, 0): %#x\n", pixels[1]);
-
-    /* Try setting the bitmap to black with a white line. */
-    glMatrixMode(GL_PROJECTION);
-    glLoadIdentity();
-    glOrtho(0.0f, 4.0f, 4.0f, 0.0f, -1.0f, 1.0f);
-    glMatrixMode(GL_MODELVIEW);
-    glLoadIdentity();
-
-    glClearColor(0.0f, 0.0f, 0.0f, 1.0f);
-    glClear(GL_COLOR_BUFFER_BIT);
-
-    glColor3f(1.0f, 1.0f, 1.0f);
-    glLineWidth(1.0f);
-    glBegin(GL_LINES);
-    glVertex2i(1, 1);
-    glVertex2i(1, 3);
-    glEnd();
-
-    glFinish();
-
-    {
-        /* Note that the line stops at (1,2) on Windows despite the second vertex being (1,3).
-         * I'm not sure if that's an implementation quirk or expected OpenGL behaviour. */
-        USHORT X = 0x7fff, _ = 0x0;
-        USHORT expected[16] = {
-            _,_,_,_,
-            _,X,_,_,
-            _,X,_,_,
-            _,_,_,_
-        };
-
-        for (int i = 0; i < 16; i++)
-        {
-            BOOL matches = (pixels[i] == expected[i]);
-            int x = i % 4;
-            int y = i / 4;
-            /* I'm using a loop so that I can put the expected image in an easy-to-understand array.
-             * Unfortunately this way of working doesn't work great with `todo_wine` since only half
-             * of the elements are a mismatch. I'm using `todo_wine_if` as a workaround. */
-            todo_wine_if(!matches) ok(matches, "Wrong color at (%d,%d). Got %#x, expected %#x\n",
-                    x, y, pixels[i], expected[i]);
-        }
-    }
-
-    /* Clean up. */
-    wglDeleteContext(gl);
-    SelectObject(hdc, old_gdi_obj);
-    DeleteObject(bitmap);
-    DeleteDC(hdc);
-}
-
-static void test_d3dkmt_rendering(void)
-{
-    static const RECT expect_rect = {0, 0, 4, 4};
-    int i, ret, count, pixel_format = 0;
-    D3DKMT_CREATEDCFROMMEMORY create;
-    D3DKMT_DESTROYDCFROMMEMORY desc;
-    UINT pixels[16 * 16], pixel;
-    GLint viewport[4], object;
-    NTSTATUS status;
-    HGLRC hglrc;
-
-    memset( (void *)pixels, 0xcd, sizeof(*pixels) * 4 * 4 );
-    create.pMemory = pixels;
-    create.Format = D3DDDIFMT_A8R8G8B8;
-    create.Width = 4;
-    create.Height = 4;
-    create.Pitch = 4 * 4;
-    create.hDeviceDc = CreateCompatibleDC( 0 );
-    create.pColorTable = NULL;
-    status = pD3DKMTCreateDCFromMemory( &create );
-    ok( !status, "got %#lx\n", status );
-    DeleteDC( create.hDeviceDc );
-    desc.hBitmap = create.hBitmap;
-    desc.hDc = create.hDc;
-
-    ret = GetPixelFormat( desc.hDc );
-    ok( ret == 0, "got %d\n", ret );
-    count = DescribePixelFormat( desc.hDc, 0, 0, NULL );
-    ok( count > 1, "got %d\n", count );
-
-    /* cannot create a GL context without a pixel format */
-
-    hglrc = wglCreateContext( desc.hDc );
-    ok( !hglrc, "wglCreateContext succeeded\n" );
-    if (hglrc) wglDeleteContext( hglrc );
-
-    /* cannot set pixel format twice */
-
-    for (i = 1; i <= count; i++)
-    {
-        PIXELFORMATDESCRIPTOR pfd = {0};
-
-        winetest_push_context( "%u", i );
-
-        ret = DescribePixelFormat( desc.hDc, i, sizeof(pfd), &pfd );
-        ok( ret == count, "got %d\n", ret );
-
-        if ((pfd.dwFlags & PFD_DRAW_TO_BITMAP) && (pfd.dwFlags & PFD_SUPPORT_OPENGL) &&
-            pfd.cColorBits == 32 && pfd.cAlphaBits > 0)
-        {
-            ret = SetPixelFormat( desc.hDc, i, &pfd );
-            if (pixel_format) ok( !ret, "SetPixelFormat succeeded\n" );
-            else ok( ret, "SetPixelFormat failed\n" );
-            if (ret) pixel_format = i;
-            ret = GetPixelFormat( desc.hDc );
-            ok( ret == pixel_format, "got %d\n", ret );
-        }
-
-        winetest_pop_context();
-    }
-
-    ok( !!pixel_format, "got pixel_format %u\n", pixel_format );
-
-    /* creating a GL context now works */
-
-    hglrc = wglCreateContext( desc.hDc );
-    ok( !!hglrc, "wglCreateContext failed, error %lu\n", GetLastError() );
-    ret = wglMakeCurrent( desc.hDc, hglrc );
-    ok( ret, "wglMakeCurrent failed, error %lu\n", GetLastError() );
-
-    glGetIntegerv( GL_READ_BUFFER, &object );
-    ok( object == GL_FRONT, "got %u\n", object );
-    glGetIntegerv( GL_DRAW_BUFFER, &object );
-    ok( object == GL_FRONT, "got %u\n", object );
-
-    memset( viewport, 0xcd, sizeof(viewport) );
-    glGetIntegerv( GL_VIEWPORT, viewport );
-    ok( EqualRect( (RECT *)viewport, &expect_rect ), "got viewport %s\n", wine_dbgstr_rect( (RECT *)viewport ) );
-
-    memset( (void *)pixels, 0xcd, sizeof(*pixels) * 4 * 4 );
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0xcdcdcd, "got %#x\n", pixel );
-
-    glClearColor( (float)0x44 / 0xff, (float)0x33 / 0xff, (float)0x22 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    ok( (pixels[0] & 0xffffff) == 0xcdcdcd, "got %#x\n", pixels[0] );
-    glFinish();
-    ok( (pixels[0] & 0xffffff) == 0x443322, "got %#x\n", pixels[0] );
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0x223344, "got %#x\n", pixel );
-
-    glClearColor( (float)0x55 / 0xff, (float)0x66 / 0xff, (float)0x77 / 0xff, (float)0x88 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    ok( (pixels[0] & 0xffffff) == 0x443322, "got %#x\n", pixels[0] );
-    glFlush();
-    ok( (pixels[0] & 0xffffff) == 0x556677, "got %#x\n", pixels[0] );
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0x776655, "got %#x\n", pixel );
-
-    glClearColor( (float)0x44 / 0xff, (float)0x33 / 0xff, (float)0x22 / 0xff, (float)0x11 / 0xff );
-    glClear( GL_COLOR_BUFFER_BIT );
-    ok( (pixels[0] & 0xffffff) == 0x556677, "got %#x\n", pixels[0] );
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0x223344, "got %#x\n", pixel );
-    ok( (pixels[0] & 0xffffff) == 0x443322, "got %#x\n", pixels[0] );
-
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( (pixel & 0xffffff) == 0x223344, "got %#x\n", pixel );
-    memset( pixels, 0xa5, sizeof(pixels) );
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    todo_wine ok( (pixel & 0xffffff) == 0xa5a5a5, "got %#x\n", pixel );
-    memset( pixels, 0xcd, sizeof(pixels) );
-
-    wglDeleteContext( hglrc );
-
-    status = pD3DKMTDestroyDCFromMemory( &desc );
-    ok( !status, "got %#lx\n", status );
-
-    winetest_pop_context();
+    }
+
+    SelectObject(hdcDst, oldDst);
+    DeleteObject(bmp2);
+    DeleteObject(bmpDst);
+    DeleteDC(hdcDst);
+    DeleteDC(hdcScreen);
 }
 
 struct wgl_thread_param
@@ -2241,10 +1148,9 @@
 {
     int res;
     HDC hread;
-    HGLRC oldctx, hglrc;
-
-    oldctx = wglGetCurrentContext();
-    if(!(hglrc = wglCreateContext(hdc)))
+    HGLRC hglrc = wglCreateContext(hdc);
+
+    if(!hglrc)
     {
         skip("wglCreateContext failed!\n");
         return;
@@ -2254,7 +1160,6 @@
     if(!res)
     {
         skip("wglMakeCurrent failed!\n");
-        wglDeleteContext(hglrc);
         return;
     }
 
@@ -2266,9 +1171,6 @@
     pwglMakeContextCurrentARB(hdc, hdc, hglrc);
     hread = pwglGetCurrentReadDCARB();
     ok(hread == hdc, "wglGetCurrentReadDCARB failed for wglMakeContextCurrent\n");
-
-    wglMakeCurrent(hdc, oldctx);
-    wglDeleteContext(hglrc);
 }
 
 static void test_dc(HWND hwnd, HDC hdc)
@@ -2530,59 +1432,6 @@
     DestroyWindow(window);
 }
 
-static DWORD CALLBACK test_window_dc_thread( void *arg )
-{
-    PIXELFORMATDESCRIPTOR pfd =
-    {
-        .nSize = sizeof(pfd),
-        .nVersion = 1,
-        .dwFlags = PFD_DRAW_TO_WINDOW | PFD_SUPPORT_OPENGL | PFD_DOUBLEBUFFER,
-        .iPixelType = PFD_TYPE_RGBA,
-        .cColorBits = 24,
-        .cDepthBits = 32,
-    };
-    HWND hwnd = arg;
-    UINT ret, pixel;
-    HGLRC ctx;
-    int format;
-    HDC hdc;
-
-    hdc = GetWindowDC( hwnd );
-    ok( hdc != NULL, "got %p\n", hdc );
-    format = ChoosePixelFormat( hdc, &pfd );
-    ok( format != 0, "got %d\n", format );
-    ret = SetPixelFormat( hdc, format, &pfd );
-    ok( ret != 0, "got %u\n", ret );
-
-    ctx = wglCreateContext( hdc );
-    ok( ctx != NULL, "got %p\n", ctx );
-    ok_ret( TRUE, wglMakeCurrent( hdc, ctx ) );
-    ok_ret( GL_NO_ERROR, glGetError() );
-    glReadBuffer( GL_BACK );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    todo_wine ok( pixel == 0xff0000ff, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    glClearColor( 0.0, 1.0, 0.0, 1.0 );
-    ok_ret( GL_NO_ERROR, glGetError() );
-    glClear( GL_COLOR_BUFFER_BIT );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( pixel == 0xff00ff00, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    ok_ret( TRUE, wglMakeCurrent( NULL, NULL ) );
-    ok_ret( TRUE, wglDeleteContext( ctx ) );
-
-    ReleaseDC( hwnd, hdc );
-    return 0;
-}
-
 static void test_window_dc(void)
 {
     PIXELFORMATDESCRIPTOR pf_desc =
@@ -2607,11 +1456,9 @@
     int pixel_format;
     HWND window;
     RECT vp, r;
-    HGLRC ctx, ctx1;
+    HGLRC ctx;
     BOOL ret;
-    HDC dc, dc1;
-    UINT pixel;
-    HANDLE thread;
+    HDC dc;
 
     window = CreateWindowA("static", "opengl32_test",
             WS_OVERLAPPEDWINDOW, 0, 0, 640, 480, 0, 0, 0, 0);
@@ -2650,112 +1497,7 @@
     ret = wglDeleteContext(ctx);
     ok(ret, "Failed to delete GL context, last error %#lx.\n", GetLastError());
 
-    ReleaseDC( window, dc );
-
-
-    dc = GetWindowDC( window );
-    ctx = wglCreateContext( dc );
-    ok( ctx != NULL, "got %p\n", ctx );
-    ok_ret( TRUE, wglMakeCurrent( dc, ctx ) );
-    glReadBuffer( GL_BACK );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    glClearColor( 1.0, 0.0, 0.0, 1.0 );
-    ok_ret( GL_NO_ERROR, glGetError() );
-    glClear( GL_COLOR_BUFFER_BIT );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( pixel == 0xff0000ff, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    ReleaseDC( window, dc );
-
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( pixel == 0xff0000ff, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    glFlush();
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( pixel == 0xff0000ff, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-
-    dc = GetWindowDC( window );
-    ok( dc != NULL, "got %p\n", dc );
-    ok_ret( TRUE, wglMakeCurrent( dc, ctx ) );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( pixel == 0xff0000ff, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    thread = CreateThread( NULL, 0, test_window_dc_thread, window, 0, NULL );
-    ok( thread != NULL, "got %p\n", thread );
-    ret = WaitForSingleObject( thread, 5000 );
-    ok( ret == 0, "got %#x\n", ret );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    todo_wine ok( pixel == 0xff00ff00, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-
-    dc1 = GetWindowDC( window );
-    ok( dc1 != NULL, "got %p\n", dc1 );
-    ok_ret( TRUE, wglMakeCurrent( dc1, ctx ) );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    todo_wine ok( pixel == 0xff00ff00, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-
-    ctx1 = wglCreateContext( dc1 );
-    ok( ctx1 != NULL, "got %p\n", ctx1 );
-    ok_ret( TRUE, wglMakeCurrent( dc1, ctx1 ) );
-    ok_ret( GL_NO_ERROR, glGetError() );
-    glReadBuffer( GL_BACK );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( pixel == 0xff00ff00, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    ok_ret( TRUE, wglMakeCurrent( NULL, NULL ) );
-    ok_ret( TRUE, wglDeleteContext( ctx1 ) );
-    ReleaseDC( window, dc1 );
-
-    ok_ret( TRUE, wglDeleteContext( ctx ) );
-    ReleaseDC( window, dc );
-
-
-    dc = GetWindowDC( window );
-    ok( dc != NULL, "got %p\n", dc );
-    ctx = wglCreateContext( dc );
-    ok( ctx != NULL, "got %p\n", ctx );
-    ok_ret( TRUE, wglMakeCurrent( dc, ctx ) );
-    ok_ret( GL_NO_ERROR, glGetError() );
-    glReadBuffer( GL_BACK );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    pixel = 0xdeadbeef;
-    glReadPixels( 0, 0, 1, 1, GL_RGBA, GL_UNSIGNED_BYTE, &pixel );
-    ok( pixel == 0xff00ff00, "got %#x\n", pixel );
-    ok_ret( GL_NO_ERROR, glGetError() );
-
-    ok_ret( TRUE, wglMakeCurrent( NULL, NULL ) );
-    ok_ret( TRUE, wglDeleteContext( ctx ) );
-    ok_ret( TRUE, SwapBuffers( dc ) );
-    ReleaseDC( window, dc );
-
+    ReleaseDC(window, dc);
     DestroyWindow(window);
 }
 
@@ -2998,6 +1740,8 @@
     DWORD err;
     HGLRC oldctx = wglGetCurrentContext();
 
+    ok(!!oldctx, "Expected to find a valid current context\n");
+
     draw_window = CreateWindowA("static", "opengl32_test",
             WS_POPUP, 0, 0, 640, 480, 0, 0, 0, 0);
     ok(!!draw_window, "Failed to create window, last error %#lx.\n", GetLastError());
@@ -3154,6 +1898,7 @@
     int interval;
 
     oldctx = wglGetCurrentContext();
+    ok(!!oldctx, "Expected to find a valid current context.\n");
 
     window1 = CreateWindowA("static", "opengl32_test",
             WS_POPUP, 0, 0, 640, 480, 0, 0, 0, 0);
@@ -3373,7 +2118,7 @@
 static void test_copy_context(HDC hdc)
 {
     HGLRC ctx, ctx2, old_ctx;
-    GLint ret;
+    BOOL ret;
 
     old_ctx = wglGetCurrentContext();
     ok(!!old_ctx, "wglGetCurrentContext failed, last error %#lx.\n", GetLastError());
@@ -3382,51 +2127,12 @@
     ok(!!ctx, "Failed to create GL context, last error %#lx.\n", GetLastError());
     ret = wglMakeCurrent(hdc, ctx);
     ok(ret, "wglMakeCurrent failed, last error %#lx.\n", GetLastError());
-
-    ret = glIsEnabled(GL_DEPTH_TEST);
-    ok(ret == 0, "got %d\n", ret);
-    glGetIntegerv(GL_DEPTH_FUNC, &ret);
-    ok(ret == GL_LESS, "got %d\n", ret);
-    glEnable(GL_DEPTH_TEST);
-    glDepthFunc(GL_GREATER);
-
     ctx2 = wglCreateContext(hdc);
     ok(!!ctx2, "Failed to create GL context, last error %#lx.\n", GetLastError());
 
-    ret = wglCopyContext(ctx, ctx2, GL_ENABLE_BIT);
+    ret = wglCopyContext(ctx, ctx2, GL_ALL_ATTRIB_BITS);
+    todo_wine
     ok(ret, "Failed to copy GL context, last error %#lx.\n", GetLastError());
-    ret = glIsEnabled(GL_DEPTH_TEST);
-    ok(ret == 1, "got %d\n", ret);
-    glGetIntegerv(GL_DEPTH_FUNC, &ret);
-    ok(ret == GL_GREATER, "got %d\n", ret);
-
-    ret = wglMakeCurrent(hdc, ctx2);
-    ok(ret, "wglMakeCurrent failed, last error %#lx.\n", GetLastError());
-    ret = glIsEnabled(GL_DEPTH_TEST);
-    ok(ret == 1, "got %d\n", ret);
-    glGetIntegerv(GL_DEPTH_FUNC, &ret);
-    ok(ret == GL_LESS, "got %d\n", ret);
-    glDepthFunc(GL_LEQUAL);
-
-    ret = wglCopyContext(ctx, ctx2, GL_ALL_ATTRIB_BITS);
-    ok(!ret, "succeeded to copy GL context.\n");
-    ok(GetLastError() == ERROR_INVALID_HANDLE, "got error %#lx.\n", GetLastError());
-
-    ret = wglMakeCurrent(hdc, ctx);
-    ok(ret, "wglMakeCurrent failed, last error %#lx.\n", GetLastError());
-    ret = glIsEnabled(GL_DEPTH_TEST);
-    ok(ret == 1, "got %d\n", ret);
-    glGetIntegerv(GL_DEPTH_FUNC, &ret);
-    ok(ret == GL_GREATER, "got %d\n", ret);
-
-    ret = wglCopyContext(ctx, ctx2, GL_ALL_ATTRIB_BITS);
-    ok(ret, "Failed to copy GL context, last error %#lx.\n", GetLastError());
-    ret = wglMakeCurrent(hdc, ctx2);
-    ok(ret, "wglMakeCurrent failed, last error %#lx.\n", GetLastError());
-    ret = glIsEnabled(GL_DEPTH_TEST);
-    ok(ret == 1, "got %d\n", ret);
-    glGetIntegerv(GL_DEPTH_FUNC, &ret);
-    ok(ret == GL_GREATER, "got %d\n", ret);
 
     ret = wglMakeCurrent(NULL, NULL);
     ok(ret, "wglMakeCurrent failed, last error %#lx.\n", GetLastError());
@@ -3486,273 +2192,6 @@
 
     ReleaseDC(child, hdc);
     DestroyWindow(child);
-}
-
-#define check_gl_error(exp) check_gl_error_(__LINE__, exp)
-static void check_gl_error_( unsigned int line, GLenum exp )
-{
-    GLenum err = glGetError();
-    ok_(__FILE__,line)( err == exp, "glGetError returned %x, expected %x\n", err, exp );
-}
-
-static void test_gl_error( HDC hdc )
-{
-    HGLRC rc, old_rc;
-    int i;
-    BOOL ret;
-
-    if (!pglDeleteSync)
-    {
-        skip( "glDeleteSync not available\n" );
-        return;
-    }
-
-    old_rc = wglGetCurrentContext();
-    rc = wglCreateContext( hdc );
-    ok( !!rc, "got %p\n", rc );
-    ret = wglMakeCurrent( hdc, rc );
-    ok( ret, "got %u\n", ret );
-
-    check_gl_error( GL_NO_ERROR );
-    glGetIntegerv( 0xdeadbeef, &i );
-    check_gl_error( GL_INVALID_ENUM );
-    check_gl_error( GL_NO_ERROR );
-
-    pglDeleteSync( (GLsync)0xdeadbeef );
-    check_gl_error( GL_INVALID_VALUE );
-    check_gl_error( GL_NO_ERROR );
-
-    glGetIntegerv( 0xdeadbeef, &i );
-    pglDeleteSync( (GLsync)0xdeadbeef );
-    check_gl_error( GL_INVALID_ENUM );
-    check_gl_error( GL_NO_ERROR );
-
-    pglDeleteSync( (GLsync)0xdeadbeef );
-    glGetIntegerv( 0xdeadbeef, &i );
-    check_gl_error( GL_INVALID_VALUE );
-    check_gl_error( GL_NO_ERROR );
-
-    ret = pglIsSync( (GLsync)0xdeadbeef );
-    ok( !ret, "glIsSync returned %x\n", ret );
-    check_gl_error( GL_NO_ERROR );
-
-    wglMakeCurrent( hdc, old_rc );
-}
-
-static void test_memory_map( HDC hdc)
-{
-    unsigned int i, major = 0, minor = 0;
-    BOOL have_persistent_storage = TRUE;
-    const char *dst_ptr, *version;
-    char *src_ptr, *ptr;
-    HGLRC rc, old_rc;
-    GLuint src, dst;
-    char data[0x1000];
-    BOOL ret;
-
-    old_rc = wglGetCurrentContext();
-
-    rc = wglCreateContext( hdc );
-    ok( !!rc, "got %p\n", rc );
-    ret = wglMakeCurrent( hdc, rc );
-    ok( ret, "got %u\n", ret );
-
-    version = (const char *)glGetString( GL_VERSION );
-    sscanf( version, "%d.%d", &major, &minor );
-    if (major < 4 || (major == 4 && minor < 4))
-    {
-        const char *extensions = (const char *)glGetString( GL_EXTENSIONS );
-        if (!extensions || !strstr( extensions, "GL_ARB_buffer_storage" ))
-        {
-            skip( "persistent map not supported\n" );
-            have_persistent_storage = FALSE;
-        }
-    }
-
-    pglGenBuffers( 1, &src );
-    pglGenBuffers( 1, &dst );
-
-    pglBindBuffer( GL_ARRAY_BUFFER, src );
-    pglBufferData( GL_ARRAY_BUFFER, sizeof(data), NULL, GL_STATIC_DRAW );
-
-    src_ptr = pglMapBuffer( GL_ARRAY_BUFFER, GL_WRITE_ONLY );
-    check_gl_error( GL_NO_ERROR );
-    ok( !((UINT_PTR)src_ptr & 0xf), "pointer not aligned\n" );
-    for (i = 0; i < sizeof(data); i++) src_ptr[i] = 'a' + i;
-
-    ptr = pglMapBuffer( GL_ARRAY_BUFFER, GL_WRITE_ONLY );
-    check_gl_error( GL_INVALID_OPERATION );
-    ok( !ptr, "repeated glMapBuffer returned %p\n", ptr );
-
-    pglUnmapBuffer( GL_ARRAY_BUFFER );
-    check_gl_error( GL_NO_ERROR );
-
-    pglUnmapBuffer( GL_ARRAY_BUFFER );
-    check_gl_error( GL_INVALID_OPERATION );
-
-    pglBindBuffer( GL_ARRAY_BUFFER, dst );
-    pglBufferData( GL_ARRAY_BUFFER, sizeof(data), NULL, GL_STATIC_DRAW );
-
-    pglBindBuffer( GL_COPY_READ_BUFFER, src );
-    pglBindBuffer( GL_COPY_WRITE_BUFFER, dst );
-    pglCopyBufferSubData( GL_COPY_READ_BUFFER, GL_COPY_WRITE_BUFFER, 0, 0, sizeof(data) );
-
-    dst_ptr = pglMapBuffer( GL_COPY_WRITE_BUFFER, GL_READ_ONLY );
-    check_gl_error( GL_NO_ERROR );
-    ok( !((UINT_PTR)dst_ptr & 0xf), "pointer not aligned\n" );
-    ok( !memcmp( dst_ptr, "abcdef", 6 ), "unexpected src data %s\n", debugstr_an(src_ptr, 6) );
-    pglUnmapBuffer( GL_COPY_WRITE_BUFFER );
-
-    if (pglMapBufferRange)
-    {
-        pglBindBuffer( GL_ARRAY_BUFFER, src );
-        src_ptr = pglMapBufferRange( GL_ARRAY_BUFFER, 3, 4, GL_MAP_READ_BIT | GL_MAP_WRITE_BIT );
-        check_gl_error( GL_NO_ERROR );
-        ok( ((UINT_PTR)src_ptr & 0xf) == 3, "pointer not aligned\n" );
-
-        ok( !memcmp( src_ptr, "defg", 4 ), "unexpected src data %s\n", debugstr_an(src_ptr, 4) );
-        for (i = 0; i < 4; i++) src_ptr[i] += 'A' - 'a';
-
-        pglUnmapBuffer( GL_ARRAY_BUFFER );
-
-        src_ptr = pglMapBufferRange( GL_ARRAY_BUFFER, 2, 10, GL_MAP_READ_BIT );
-        ok( ((UINT_PTR)src_ptr & 0xf) == 2, "pointer not aligned\n" );
-
-        ok( !memcmp( src_ptr, "cDEFGhijkl", 10 ), "unexpected src data %s\n", debugstr_an(src_ptr, 10) );
-
-        pglUnmapBuffer( GL_ARRAY_BUFFER );
-
-        pglCopyBufferSubData( GL_COPY_READ_BUFFER, GL_COPY_WRITE_BUFFER, 0, 0, sizeof(data) );
-
-        pglBindBuffer( GL_ARRAY_BUFFER, dst );
-        dst_ptr = pglMapBufferRange( GL_ARRAY_BUFFER, 2, 10, GL_MAP_READ_BIT );
-        ok( ((UINT_PTR)dst_ptr & 0xf) == 2, "pointer not aligned\n" );
-
-        ok( !memcmp( dst_ptr, "cDEFGhijkl", 10 ), "unexpected src data %s\n", debugstr_an(dst_ptr, 10) );
-
-        pglUnmapBuffer( GL_ARRAY_BUFFER );
-        check_gl_error( GL_NO_ERROR );
-    }
-    else skip( "glMapBufferRange not available\n" );
-
-    if (have_persistent_storage)
-    {
-        for (i = 0; i < sizeof(data); i++) data[i] = '0' + i;
-        pglBufferStorage( GL_COPY_READ_BUFFER, sizeof(data), data,
-                          GL_MAP_READ_BIT | GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
-
-        pglBufferStorage( GL_COPY_WRITE_BUFFER, sizeof(data), NULL,
-                          GL_MAP_READ_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
-
-        src_ptr = pglMapBufferRange( GL_COPY_READ_BUFFER, 2, sizeof(data) - 2,
-                                     GL_MAP_READ_BIT | GL_MAP_WRITE_BIT | GL_MAP_FLUSH_EXPLICIT_BIT | GL_MAP_PERSISTENT_BIT );
-        ok( ((UINT_PTR)src_ptr & 0xf) == 2, "pointer not aligned\n" );
-
-        dst_ptr = pglMapBufferRange( GL_COPY_WRITE_BUFFER, 0, sizeof(data),
-                                     GL_MAP_READ_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
-        ok( ((UINT_PTR)dst_ptr & 0xf) == 0, "pointer not aligned\n" );
-
-        ok( src_ptr[0] == '2', "src_ptr[0] = %x (%c)\n", src_ptr[0], src_ptr[0] );
-        src_ptr[0] += 'a' - '0';
-        pglFlushMappedBufferRange( GL_COPY_READ_BUFFER, 0, 16 );
-
-        pglCopyBufferSubData( GL_COPY_READ_BUFFER, GL_COPY_WRITE_BUFFER, 0, 0, sizeof(data) );
-        glFinish();
-        ok( !memcmp( dst_ptr, "01c3456789", 8 ), "unexpected dst data %s\n", debugstr_an(dst_ptr, 10) );
-
-        src_ptr[1] += 'A' - '0';
-        pglCopyBufferSubData( GL_COPY_READ_BUFFER, GL_COPY_WRITE_BUFFER, 0, 0, sizeof(data) );
-        glFinish();
-        ok( !memcmp( dst_ptr, "01cD456789", 8 ), "unexpected dst data %s\n", debugstr_an(dst_ptr, 10) );
-
-        pglUnmapBuffer( GL_COPY_WRITE_BUFFER );
-        pglUnmapBuffer( GL_COPY_READ_BUFFER );
-        check_gl_error( GL_NO_ERROR );
-    }
-
-    pglDeleteBuffers( 1, &src );
-    pglDeleteBuffers( 1, &dst );
-
-    if (major > 4 || (major == 4 && minor >= 5))
-    {
-        pglCreateBuffers( 1, &src );
-        pglCreateBuffers( 1, &dst );
-        check_gl_error( GL_NO_ERROR );
-
-        pglNamedBufferData( src, 0x1000, NULL, GL_STATIC_DRAW );
-        check_gl_error( GL_NO_ERROR );
-
-        src_ptr = pglMapNamedBuffer( src, GL_WRITE_ONLY );
-        check_gl_error( GL_NO_ERROR );
-        ok( !((UINT_PTR)src_ptr & 0xf), "pointer not aligned\n" );
-        for (i = 0; i < 0x1000; i++) src_ptr[i] = 'a' + i;
-
-        ptr = pglMapNamedBuffer( src, GL_WRITE_ONLY );
-        check_gl_error( GL_INVALID_OPERATION );
-        ok( !ptr, "repeated glMapBuffer returned %p\n", ptr );
-
-        pglUnmapNamedBuffer( src );
-        check_gl_error( GL_NO_ERROR );
-
-        pglUnmapNamedBuffer( src );
-        check_gl_error( GL_INVALID_OPERATION );
-
-        pglNamedBufferData( dst, 0x1000, NULL, GL_STATIC_DRAW );
-
-        pglCopyNamedBufferSubData( src, dst, 0, 0, 0x1000 );
-
-        dst_ptr = pglMapNamedBuffer( dst, GL_READ_ONLY );
-        check_gl_error( GL_NO_ERROR );
-        ok( !((UINT_PTR)dst_ptr & 0xf), "pointer not aligned\n" );
-        ok( !memcmp( dst_ptr, "abcdef", 6 ), "unexpected src data %s\n", debugstr_an(src_ptr, 6) );
-        pglUnmapNamedBuffer( dst );
-
-        pglDeleteBuffers( 1, &src );
-        pglDeleteBuffers( 1, &dst );
-
-        pglCreateBuffers( 1, &src );
-        pglCreateBuffers( 1, &dst );
-
-        for (i = 0; i < sizeof(data); i++) data[i] = '0' + i;
-        pglNamedBufferStorage( src, sizeof(data), data,
-                               GL_MAP_READ_BIT | GL_MAP_WRITE_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
-
-        pglNamedBufferStorage( dst, sizeof(data), NULL,
-                               GL_MAP_READ_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
-
-        src_ptr = pglMapNamedBufferRange( src, 2, sizeof(data) - 2,
-                                          GL_MAP_READ_BIT | GL_MAP_WRITE_BIT | GL_MAP_FLUSH_EXPLICIT_BIT | GL_MAP_PERSISTENT_BIT );
-        ok( ((UINT_PTR)src_ptr & 0xf) == 2, "pointer not aligned\n" );
-
-        dst_ptr = pglMapNamedBufferRange( dst, 0, sizeof(data),
-                                          GL_MAP_READ_BIT | GL_MAP_PERSISTENT_BIT | GL_MAP_COHERENT_BIT );
-        check_gl_error( GL_NO_ERROR );
-        ok( ((UINT_PTR)dst_ptr & 0xf) == 0, "pointer not aligned\n" );
-
-        ok( src_ptr[0] == '2', "src_ptr[0] = %x (%c)\n", src_ptr[0], src_ptr[0] );
-        src_ptr[0] += 'a' - '0';
-        pglFlushMappedNamedBufferRange( src, 0, 16 );
-        check_gl_error( GL_NO_ERROR );
-
-        pglCopyNamedBufferSubData( src, dst, 0, 0, sizeof(data) );
-        glFinish();
-        ok( !memcmp( dst_ptr, "01c3456789", 8 ), "unexpected dst data %s\n", debugstr_an(dst_ptr, 10) );
-
-        src_ptr[1] += 'A' - '0';
-        pglCopyNamedBufferSubData( src, dst, 0, 0, sizeof(data) );
-        glFinish();
-        ok( !memcmp( dst_ptr, "01cD456789", 8 ), "unexpected dst data %s\n", debugstr_an(dst_ptr, 10) );
-
-        pglUnmapNamedBuffer( src );
-        pglUnmapNamedBuffer( dst );
-
-        pglDeleteBuffers( 1, &src );
-        pglDeleteBuffers( 1, &dst );
-        check_gl_error( GL_NO_ERROR );
-    }
-    else skip( "Named buffers not supported by OpenGL %s\n", version );
-
-    wglMakeCurrent( hdc, old_rc );
 }
 
 START_TEST(opengl)
@@ -3784,16 +2223,10 @@
     ok(hwnd != NULL, "err: %ld\n", GetLastError());
     if (hwnd)
     {
-        HMODULE gdi32 = GetModuleHandleA("gdi32.dll");
         HDC hdc;
         int iPixelFormat, res;
         HGLRC hglrc;
         DWORD error;
-
-        pD3DKMTCreateDCFromMemory  = (void *)GetProcAddress( gdi32, "D3DKMTCreateDCFromMemory" );
-        pD3DKMTDestroyDCFromMemory = (void *)GetProcAddress( gdi32, "D3DKMTDestroyDCFromMemory" );
-
-        check_gl_error( GL_INVALID_OPERATION );
         ShowWindow(hwnd, SW_SHOW);
 
         hdc = GetDC(hwnd);
@@ -3817,8 +2250,6 @@
 
         test_bitmap_rendering( TRUE );
         test_bitmap_rendering( FALSE );
-        test_16bit_bitmap_rendering();
-        test_d3dkmt_rendering();
         test_minimized();
         test_window_dc();
         test_message_window();
@@ -3872,8 +2303,6 @@
         test_gdi_dbuf(hdc);
         test_acceleration(hdc);
         test_framebuffer();
-        test_memory_map(hdc);
-        test_gl_error(hdc);
 
         wgl_extensions = pwglGetExtensionsStringARB(hdc);
         if(wgl_extensions == NULL) skip("Skipping opengl32 tests because this OpenGL implementation doesn't support WGL extensions!\n");
