diff --git a/dlls/mf/tests/mf.c b/dlls/mf/tests/mf.c
--- a/dlls/mf/tests/mf.c
+++ b/dlls/mf/tests/mf.c
@@ -43,9 +43,6 @@
 #include "devpkey.h"
 #include "evr9.h"
 
-#include "initguid.h"
-#include "netlistmgr.h"
-
 #define DEFINE_EXPECT(func) \
     static BOOL expect_ ## func = FALSE, called_ ## func = FALSE
 
@@ -88,8 +85,6 @@
 HRESULT (WINAPI *pMFCreateMediaBufferFromMediaType)(IMFMediaType *media_type, LONGLONG duration, DWORD min_length,
         DWORD min_alignment, IMFMediaBuffer **buffer);
 BOOL has_video_processor;
-
-DEFINE_GUID(InvalidServiceGUID, 0x12345678, 0x1234, 0x5678, 0x12, 0x34, 0x11, 0x11, 0x22, 0x22, 0x33, 0x33);
 
 static BOOL is_vista(void)
 {
@@ -578,7 +573,6 @@
 struct test_handler
 {
     IMFMediaTypeHandler IMFMediaTypeHandler_iface;
-    LONG refcount;
 
     ULONG set_current_count;
     IMFMediaType *current_type;
@@ -609,25 +603,12 @@
 
 static ULONG WINAPI test_handler_AddRef(IMFMediaTypeHandler *iface)
 {
-    struct test_handler *impl = impl_from_IMFMediaTypeHandler(iface);
-    return InterlockedIncrement(&impl->refcount);
+    return 2;
 }
 
 static ULONG WINAPI test_handler_Release(IMFMediaTypeHandler *iface)
 {
-    struct test_handler *impl = impl_from_IMFMediaTypeHandler(iface);
-    ULONG refcount = InterlockedDecrement(&impl->refcount);
-
-    if (!refcount)
-    {
-        if (impl->current_type)
-            IMFMediaType_Release(impl->current_type);
-        /* references to invalid_type and media_types are not held. */
-
-        free(impl);
-    }
-
-    return refcount;
+    return 1;
 }
 
 static HRESULT WINAPI test_handler_IsMediaTypeSupported(IMFMediaTypeHandler *iface, IMFMediaType *in_type,
@@ -734,33 +715,11 @@
     test_handler_GetMajorType,
 };
 
-static struct test_handler *create_test_handler(void)
-{
-    struct test_handler *handler;
-
-    handler = calloc(1, sizeof(*handler));
-    handler->IMFMediaTypeHandler_iface.lpVtbl = &test_handler_vtbl;
-    handler->refcount = 1;
-
-    return handler;
-};
-
-static void test_handler_clear_current_type(struct test_handler *handler)
-{
-    if (handler->current_type)
-    {
-        IMFMediaType_Release(handler->current_type);
-        handler->current_type = NULL;
-    }
-}
+static const struct test_handler test_handler = {.IMFMediaTypeHandler_iface.lpVtbl = &test_handler_vtbl};
 
 struct test_media_sink
 {
     IMFMediaSink IMFMediaSink_iface;
-    LONG refcount;
-    IMFMediaTypeHandler *handler;
-    IMFPresentationClock *clock;
-    IMFStreamSink *stream;
     BOOL shutdown;
 };
 
@@ -784,25 +743,12 @@
 
 static ULONG WINAPI test_media_sink_AddRef(IMFMediaSink *iface)
 {
-    struct test_media_sink *sink = impl_from_IMFMediaSink(iface);
-    return InterlockedIncrement(&sink->refcount);
+    return 2;
 }
 
 static ULONG WINAPI test_media_sink_Release(IMFMediaSink *iface)
 {
-    struct test_media_sink *sink = impl_from_IMFMediaSink(iface);
-    ULONG refcount = InterlockedDecrement(&sink->refcount);
-
-    if (!refcount)
-    {
-        if (!sink->shutdown)
-            IMFMediaSink_Shutdown(iface);
-        if (sink->handler)
-            IMFMediaTypeHandler_Release(sink->handler);
-        free(sink);
-    }
-
-    return refcount;
+    return 1;
 }
 
 static HRESULT WINAPI test_media_sink_GetCharacteristics(IMFMediaSink *iface, DWORD *characteristics)
@@ -824,147 +770,43 @@
     return E_NOTIMPL;
 }
 
-DEFINE_EXPECT(test_media_sink_GetStreamSinkCount);
-DEFINE_EXPECT(test_media_sink_SetPresentationClock);
-DEFINE_EXPECT(test_media_sink_GetPresentationClock);
-
 static HRESULT WINAPI test_media_sink_GetStreamSinkCount(IMFMediaSink *iface, DWORD *count)
 {
-    HRESULT hr;
-    if (expect_test_media_sink_GetStreamSinkCount)
-    {
-        *count = 1;
-        hr = S_OK;
-    }
-    else
-    {
-        hr = E_NOTIMPL;
-    }
-    CHECK_EXPECT(test_media_sink_GetStreamSinkCount);
-    return hr;
-}
-
-static HRESULT WINAPI test_media_sink_GetStreamSinkByIndex(IMFMediaSink *iface, DWORD index, IMFStreamSink **sink)
-{
-    struct test_media_sink *sink_impl = impl_from_IMFMediaSink(iface);
-    if (!index && sink_impl->stream)
-    {
-        IMFStreamSink_AddRef(*sink = sink_impl->stream);
-        return S_OK;
-    }
     ok(0, "Unexpected call.\n");
     return E_NOTIMPL;
 }
 
-static HRESULT WINAPI test_media_sink_GetStreamSinkById(IMFMediaSink *iface, DWORD stream_sink_id, IMFStreamSink **sink)
+static HRESULT WINAPI test_media_sink_GetStreamSinkByIndex(IMFMediaSink *iface, DWORD index, IMFStreamSink **sink)
 {
     ok(0, "Unexpected call.\n");
     return E_NOTIMPL;
 }
 
+static HRESULT WINAPI test_media_sink_GetStreamSinkById(IMFMediaSink *iface, DWORD stream_sink_id, IMFStreamSink **sink)
+{
+    ok(0, "Unexpected call.\n");
+    return E_NOTIMPL;
+}
+
 static HRESULT WINAPI test_media_sink_SetPresentationClock(IMFMediaSink *iface, IMFPresentationClock *clock)
 {
+    ok(0, "Unexpected call.\n");
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI test_media_sink_GetPresentationClock(IMFMediaSink *iface, IMFPresentationClock **clock)
+{
+    ok(0, "Unexpected call.\n");
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI test_media_sink_Shutdown(IMFMediaSink *iface)
+{
     struct test_media_sink *sink = impl_from_IMFMediaSink(iface);
-    HRESULT hr;
-
-    if (expect_test_media_sink_SetPresentationClock)
-    {
-        if (sink->clock) IMFPresentationClock_Release(clock);
-        IMFPresentationClock_AddRef(sink->clock = clock);
-        hr = S_OK;
-    }
-    else
-    {
-        hr = E_NOTIMPL;
-    }
-    CHECK_EXPECT(test_media_sink_SetPresentationClock);
-    return hr;
-}
-
-static HRESULT WINAPI test_media_sink_GetPresentationClock(IMFMediaSink *iface, IMFPresentationClock **clock)
-{
-    struct test_media_sink *sink = impl_from_IMFMediaSink(iface);
-
-    CHECK_EXPECT2(test_media_sink_GetPresentationClock);
-    if (expect_test_media_sink_GetPresentationClock)
-    {
-        if (sink->clock)
-        {
-            IMFPresentationClock_AddRef(*clock = sink->clock);
-            return S_OK;
-        }
-        return MF_E_NO_CLOCK;
-    }
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_media_sink_Shutdown(IMFMediaSink *iface)
-{
-    struct test_media_sink *sink = impl_from_IMFMediaSink(iface);
-
-    if (sink->clock)
-    {
-        IMFPresentationClock_Release(sink->clock);
-        sink->clock = NULL;
-    }
-    if (sink->stream)
-    {
-        IMFStreamSink_Release(sink->stream);
-        sink->stream = NULL;
-    }
-
     ok(!sink->shutdown, "Unexpected call.\n");
     sink->shutdown = TRUE;
     return S_OK;
 }
-
-enum object_state
-{
-    SOURCE_START,
-    SOURCE_PAUSE,
-    SOURCE_STOP,
-    SOURCE_SHUTDOWN,
-    SOURCE_REQUEST_SAMPLE,
-    SINK_ON_CLOCK_START,
-    SINK_ON_CLOCK_PAUSE,
-    SINK_ON_CLOCK_STOP,
-    SINK_ON_CLOCK_RESTART,
-    SINK_ON_CLOCK_SETRATE,
-    SINK_FLUSH,
-    SINK_PROCESS_SAMPLE,
-    MFT_BEGIN,
-    MFT_START,
-    MFT_FLUSH,
-    MFT_PROCESS_INPUT,
-    MFT_PROCESS_OUTPUT,
-};
-
-#define MAX_OBJECT_STATE 1024
-
-struct object_state_record
-{
-    enum object_state states[MAX_OBJECT_STATE];
-    unsigned int state_count;
-};
-static struct object_state_record actual_object_state_record;
-
-#define add_object_state(a, b) _add_object_state(__LINE__, a, b)
-static void _add_object_state(int line, struct object_state_record *record, enum object_state state)
-{
-    ok_(__FILE__, line)(record->state_count < MAX_OBJECT_STATE, "exceeded state_count maximum %d.\n", MAX_OBJECT_STATE);
-    if (record->state_count < MAX_OBJECT_STATE)
-        record->states[record->state_count++] = state;
-}
-
-#define compare_object_states(a, b) _compare_object_states(__LINE__, a, b)
-static void _compare_object_states(int line, const struct object_state_record *r1,
-        const struct object_state_record *r2)
-{
-    ok_(__FILE__, line)(r1->state_count == r2->state_count, "State count not equal.\n");
-    if (r1->state_count == r2->state_count)
-        ok_(__FILE__, line)(!memcmp(r1->states, r2->states, sizeof(enum object_state) * r1->state_count), "Got different states.\n");
-}
-
 
 static const IMFMediaSinkVtbl test_media_sink_vtbl =
 {
@@ -982,18 +824,17 @@
     test_media_sink_Shutdown,
 };
 
+static const struct test_media_sink test_media_sink = {.IMFMediaSink_iface.lpVtbl = &test_media_sink_vtbl};
+
 struct test_stream_sink
 {
     IMFStreamSink IMFStreamSink_iface;
     IMFGetService IMFGetService_iface;
-    LONG refcount;
     IMFMediaTypeHandler *handler;
     IMFMediaSink *media_sink;
 
     IMFAttributes *attributes;
     IUnknown *device_manager;
-
-    IMFMediaEventQueue *event_queue;
 };
 
 static struct test_stream_sink *impl_from_IMFStreamSink(IMFStreamSink *iface)
@@ -1032,30 +873,12 @@
 
 static ULONG WINAPI test_stream_sink_AddRef(IMFStreamSink *iface)
 {
-    struct test_stream_sink *sink = impl_from_IMFStreamSink(iface);
-    return InterlockedIncrement(&sink->refcount);
+    return 2;
 }
 
 static ULONG WINAPI test_stream_sink_Release(IMFStreamSink *iface)
 {
-    struct test_stream_sink *sink = impl_from_IMFStreamSink(iface);
-    ULONG refcount = InterlockedDecrement(&sink->refcount);
-
-    if (!refcount)
-    {
-        if (sink->handler)
-            IMFMediaTypeHandler_Release(sink->handler);
-        if (sink->media_sink)
-            IMFMediaSink_Release(sink->media_sink);
-        if (sink->event_queue)
-        {
-            IMFMediaEventQueue_Shutdown(sink->event_queue);
-            IMFMediaEventQueue_Release(sink->event_queue);
-        }
-        free(sink);
-    }
-
-    return refcount;
+    return 1;
 }
 
 static HRESULT WINAPI test_stream_sink_GetEvent(IMFStreamSink *iface, DWORD flags, IMFMediaEvent **event)
@@ -1066,11 +889,6 @@
 
 static HRESULT WINAPI test_stream_sink_BeginGetEvent(IMFStreamSink *iface, IMFAsyncCallback *callback, IUnknown *state)
 {
-    struct test_stream_sink *sink = impl_from_IMFStreamSink(iface);
-
-    if (sink->event_queue)
-        return IMFMediaEventQueue_BeginGetEvent(sink->event_queue, callback, state);
-
     ok(0, "Unexpected call.\n");
     return E_NOTIMPL;
 }
@@ -1078,11 +896,6 @@
 static HRESULT WINAPI test_stream_sink_EndGetEvent(IMFStreamSink *iface, IMFAsyncResult *result,
         IMFMediaEvent **event)
 {
-    struct test_stream_sink *sink = impl_from_IMFStreamSink(iface);
-
-    if (sink->event_queue)
-        return IMFMediaEventQueue_EndGetEvent(sink->event_queue, result, event);
-
     ok(0, "Unexpected call.\n");
     return E_NOTIMPL;
 }
@@ -1090,11 +903,6 @@
 static HRESULT WINAPI test_stream_sink_QueueEvent(IMFStreamSink *iface, MediaEventType event_type,
         REFGUID ext_type, HRESULT hr, const PROPVARIANT *value)
 {
-    struct test_stream_sink *sink = impl_from_IMFStreamSink(iface);
-
-    if (sink->event_queue)
-        return IMFMediaEventQueue_QueueEventParamVar(sink->event_queue, event_type, ext_type, hr, value);
-
     ok(0, "Unexpected call.\n");
     return E_NOTIMPL;
 }
@@ -1134,22 +942,10 @@
     return E_NOTIMPL;
 }
 
-DEFINE_EXPECT(test_stream_sink_ProcessSample);
-DEFINE_EXPECT(test_stream_sink_Flush);
-
 static HRESULT WINAPI test_stream_sink_ProcessSample(IMFStreamSink *iface, IMFSample *sample)
 {
-    HRESULT hr;
-
-    if (expect_test_stream_sink_ProcessSample)
-        hr = S_OK;
-    else
-        hr = E_NOTIMPL;
-
-    CHECK_EXPECT(test_stream_sink_ProcessSample);
-    add_object_state(&actual_object_state_record, SINK_PROCESS_SAMPLE);
-
-    return hr;
+    ok(0, "Unexpected call.\n");
+    return E_NOTIMPL;
 }
 
 static HRESULT WINAPI test_stream_sink_PlaceMarker(IMFStreamSink *iface, MFSTREAMSINK_MARKER_TYPE marker_type,
@@ -1161,17 +957,8 @@
 
 static HRESULT WINAPI test_stream_sink_Flush(IMFStreamSink *iface)
 {
-    HRESULT hr;
-
-    if (expect_test_stream_sink_Flush)
-        hr = S_OK;
-    else
-        hr = E_NOTIMPL;
-
-    CHECK_EXPECT(test_stream_sink_Flush);
-    add_object_state(&actual_object_state_record, SINK_FLUSH);
-
-    return hr;
+    ok(0, "Unexpected call.\n");
+    return E_NOTIMPL;
 }
 
 static const IMFStreamSinkVtbl test_stream_sink_vtbl =
@@ -1232,47 +1019,11 @@
     test_stream_sink_get_service_GetService,
 };
 
-static struct test_stream_sink *create_test_stream_sink(IMFMediaSink *media_sink,
-        IMFMediaTypeHandler *handler, BOOL create_queue)
-{
-    struct test_stream_sink *sink;
-
-    sink = calloc(1, sizeof(*sink));
-    sink->IMFStreamSink_iface.lpVtbl = &test_stream_sink_vtbl,
-    sink->IMFGetService_iface.lpVtbl = &test_stream_sink_get_service_vtbl,
-    sink->refcount = 1;
-    if (handler)
-        IMFMediaTypeHandler_AddRef(sink->handler = handler);
-    if (media_sink)
-        IMFMediaSink_AddRef(sink->media_sink = media_sink);
-    if (create_queue)
-        MFCreateEventQueue(&sink->event_queue);
-
-    return sink;
-}
-
-static void reset_test_media_sink(struct test_media_sink *sink)
-{
-    if (sink->shutdown)
-    {
-        sink->stream = &create_test_stream_sink(&sink->IMFMediaSink_iface, sink->handler, TRUE)->IMFStreamSink_iface;
-        sink->shutdown = FALSE;
-    }
-}
-
-static struct test_media_sink *create_test_media_sink(IMFMediaTypeHandler *handler)
-{
-    struct test_media_sink *sink;
-
-    sink = calloc(1, sizeof(*sink));
-    sink->IMFMediaSink_iface.lpVtbl = &test_media_sink_vtbl;
-    sink->refcount = 1;
-    if (handler)
-        IMFMediaTypeHandler_AddRef(sink->handler = handler);
-    sink->stream = &create_test_stream_sink(&sink->IMFMediaSink_iface, handler, TRUE)->IMFStreamSink_iface;
-
-    return sink;
-}
+static const struct test_stream_sink test_stream_sink =
+{
+    .IMFStreamSink_iface.lpVtbl = &test_stream_sink_vtbl,
+    .IMFGetService_iface.lpVtbl = &test_stream_sink_get_service_vtbl,
+};
 
 struct test_callback
 {
@@ -1282,9 +1033,6 @@
     HANDLE event;
     IMFMediaEvent *media_event;
     BOOL check_media_event;
-
-    CRITICAL_SECTION cs;
-    BOOL subscribed;
 };
 
 static struct test_callback *impl_from_IMFAsyncCallback(IMFAsyncCallback *iface)
@@ -1322,7 +1070,6 @@
         if (callback->media_event)
             IMFMediaEvent_Release(callback->media_event);
         CloseHandle(callback->event);
-        DeleteCriticalSection(&callback->cs);
         free(callback);
     }
 
@@ -1353,23 +1100,13 @@
 
         hr = IMFAsyncResult_GetState(result, &object);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-        EnterCriticalSection(&callback->cs);
-
         hr = IMFMediaEventGenerator_EndGetEvent((IMFMediaEventGenerator *)object,
                 result, &callback->media_event);
-        callback->subscribed = FALSE;
-        SetEvent(callback->event);
-
-        LeaveCriticalSection(&callback->cs);
-
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
         IUnknown_Release(object);
     }
-    else
-    {
-        SetEvent(callback->event);
-    }
+
+    SetEvent(callback->event);
 
     return S_OK;
 }
@@ -1393,15 +1130,14 @@
     callback->refcount = 1;
     callback->check_media_event = check_media_event;
     callback->IMFAsyncCallback_iface.lpVtbl = &testcallbackvtbl;
-    InitializeCriticalSection(&callback->cs);
     callback->event = CreateEventW(NULL, FALSE, FALSE, NULL);
     ok(!!callback->event, "CreateEventW failed, error %lu\n", GetLastError());
 
     return &callback->IMFAsyncCallback_iface;
 }
 
-#define gen_wait_media_event(a, b, c, d, e) gen_wait_media_event_(__LINE__, a, b, c, d, e)
-static HRESULT gen_wait_media_event_(int line, IMFMediaEventGenerator *event_generator, IMFAsyncCallback *callback,
+#define wait_media_event(a, b, c, d, e) wait_media_event_(__LINE__, a, b, c, d, e)
+static HRESULT wait_media_event_(int line, IMFMediaSession *session, IMFAsyncCallback *callback,
         MediaEventType expect_type, DWORD timeout, PROPVARIANT *value)
 {
     struct test_callback *impl = impl_from_IMFAsyncCallback(callback);
@@ -1412,22 +1148,9 @@
 
     do
     {
-        ret = WAIT_TIMEOUT;
-        EnterCriticalSection(&impl->cs);
-        if (!impl->subscribed && (ret = WaitForSingleObject(impl->event, 0)) != WAIT_OBJECT_0)
-        {
-            impl->subscribed = TRUE;
-            hr = IMFMediaEventGenerator_BeginGetEvent(event_generator, &impl->IMFAsyncCallback_iface, (IUnknown *)event_generator);
-            LeaveCriticalSection(&impl->cs);
-            ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-            ret = WaitForSingleObject(impl->event, timeout);
-        }
-        else
-        {
-            LeaveCriticalSection(&impl->cs);
-            if (ret != WAIT_OBJECT_0)
-                ret = WaitForSingleObject(impl->event, timeout);
-        }
+        hr = IMFMediaSession_BeginGetEvent(session, &impl->IMFAsyncCallback_iface, (IUnknown *)session);
+        ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+        ret = WaitForSingleObject(impl->event, timeout);
         ok_(__FILE__, line)(ret == WAIT_OBJECT_0, "WaitForSingleObject returned %lu\n", ret);
         hr = IMFMediaEvent_GetType(impl->media_event, &type);
         ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -1448,15 +1171,8 @@
     return status;
 }
 
-#define wait_media_event(a, b, c, d, e) wait_media_event_(__LINE__, a, b, c, d, e)
-static HRESULT wait_media_event_(int line, IMFMediaSession *session, IMFAsyncCallback *callback,
-        MediaEventType expect_type, DWORD timeout, PROPVARIANT *value)
-{
-    return gen_wait_media_event_(line, (IMFMediaEventGenerator*) session, callback, expect_type, timeout, value);
-}
-
-#define gen_wait_media_event_until_blocking(a, b, c, d, e) gen_wait_media_event_until_blocking_(__LINE__, a, b, c, d, e)
-static HRESULT gen_wait_media_event_until_blocking_(int line, IMFMediaEventGenerator *event_generator, IMFAsyncCallback *callback,
+#define wait_media_event_until_blocking(a, b, c, d, e) wait_media_event_until_blocking_(__LINE__, a, b, c, d, e)
+static HRESULT wait_media_event_until_blocking_(int line, IMFMediaSession *session, IMFAsyncCallback *callback,
                                  MediaEventType expect_type, DWORD timeout, PROPVARIANT *value)
 {
     struct test_callback *impl = impl_from_IMFAsyncCallback(callback);
@@ -1467,26 +1183,10 @@
 
     do
     {
-        ret = WAIT_TIMEOUT;
-        EnterCriticalSection(&impl->cs);
-        if (!impl->subscribed && (ret = WaitForSingleObject(impl->event, 0)) != WAIT_OBJECT_0)
-        {
-            impl->subscribed = TRUE;
-            hr = IMFMediaEventGenerator_BeginGetEvent(event_generator, &impl->IMFAsyncCallback_iface, (IUnknown *)event_generator);
-            LeaveCriticalSection(&impl->cs);
-            ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-            ret = WaitForSingleObject(impl->event, timeout);
-        }
-        else
-        {
-            LeaveCriticalSection(&impl->cs);
-            if (ret != WAIT_OBJECT_0)
-                ret = WaitForSingleObject(impl->event, timeout);
-        }
-
-        if (ret == WAIT_TIMEOUT)
-            return WAIT_TIMEOUT;
-
+        hr = IMFMediaSession_BeginGetEvent(session, &impl->IMFAsyncCallback_iface, (IUnknown *)session);
+        ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+        ret = WaitForSingleObject(impl->event, timeout);
+        if (ret == WAIT_TIMEOUT) return WAIT_TIMEOUT;
         hr = IMFMediaEvent_GetType(impl->media_event, &type);
         ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     } while (type != expect_type);
@@ -1504,13 +1204,6 @@
     ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
     return status;
-}
-
-#define wait_media_event_until_blocking(a, b, c, d, e) wait_media_event_until_blocking_(__LINE__, a, b, c, d, e)
-static HRESULT wait_media_event_until_blocking_(int line, IMFMediaSession *session, IMFAsyncCallback *callback,
-                                 MediaEventType expect_type, DWORD timeout, PROPVARIANT *value)
-{
-    return gen_wait_media_event_until_blocking_(line, (IMFMediaEventGenerator*) session, callback, expect_type, timeout, value);
 }
 
 static IMFMediaSource *create_media_source(const WCHAR *name, const WCHAR *mime)
@@ -1559,6 +1252,45 @@
     return source;
 }
 
+enum object_state
+{
+    SOURCE_START,
+    SOURCE_PAUSE,
+    SOURCE_STOP,
+    SOURCE_SHUTDOWN,
+    SINK_ON_CLOCK_START,
+    SINK_ON_CLOCK_PAUSE,
+    SINK_ON_CLOCK_STOP,
+    SINK_ON_CLOCK_RESTART,
+    SINK_ON_CLOCK_SETRATE,
+};
+
+#define MAX_OBJECT_STATE 1024
+
+struct object_state_record
+{
+    enum object_state states[MAX_OBJECT_STATE];
+    unsigned int state_count;
+};
+static struct object_state_record actual_object_state_record;
+
+#define add_object_state(a, b) _add_object_state(__LINE__, a, b)
+static void _add_object_state(int line, struct object_state_record *record, enum object_state state)
+{
+    ok_(__FILE__, line)(record->state_count < MAX_OBJECT_STATE, "exceeded state_count maximum %d.\n", MAX_OBJECT_STATE);
+    if (record->state_count < MAX_OBJECT_STATE)
+        record->states[record->state_count++] = state;
+}
+
+#define compare_object_states(a, b) _compare_object_states(__LINE__, a, b)
+static void _compare_object_states(int line, const struct object_state_record *r1,
+        const struct object_state_record *r2)
+{
+    ok_(__FILE__, line)(r1->state_count == r2->state_count, "State count not equal.\n");
+    if (r1->state_count == r2->state_count)
+        ok_(__FILE__, line)(!memcmp(r1->states, r2->states, sizeof(enum object_state) * r1->state_count), "Got different states.\n");
+}
+
 enum source_state
 {
     SOURCE_STOPPED,
@@ -1575,9 +1307,6 @@
     LONGLONG sample_duration;
     LONGLONG sample_time;
     BOOL is_new;
-    BOOL test_expect;
-    BOOL delay_sample;
-    IMFSample *delayed_sample;
     LONG refcount;
 };
 
@@ -1617,8 +1346,6 @@
 
     if (!refcount)
     {
-        if (stream->delayed_sample)
-            IMFSample_Release(stream->delayed_sample);
         IMFMediaEventQueue_Release(stream->event_queue);
         free(stream);
     }
@@ -1670,8 +1397,6 @@
 
     return S_OK;
 }
-
-DEFINE_EXPECT(test_media_stream_RequestSample);
 
 static HRESULT WINAPI test_media_stream_RequestSample(IMFMediaStream *iface, IUnknown *token)
 {
@@ -1680,20 +1405,6 @@
     IMFSample *sample;
     HRESULT hr;
 
-    if (stream->test_expect)
-    {
-        if (expect_test_media_stream_RequestSample)
-            hr = S_OK;
-        else
-            hr = E_NOTIMPL;
-
-        CHECK_EXPECT(test_media_stream_RequestSample);
-        add_object_state(&actual_object_state_record, SOURCE_REQUEST_SAMPLE);
-
-        if (FAILED(hr))
-            return hr;
-    }
-
     hr = MFCreateSample(&sample);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     if (stream->sample_duration)
@@ -1725,18 +1436,10 @@
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     IMFMediaBuffer_Release(buffer);
 
-    if (stream->delay_sample)
-    {
-        if (stream->delayed_sample) IMFSample_Release(stream->delayed_sample);
-        stream->delayed_sample = sample;
-    }
-    else
-    {
-        hr = IMFMediaEventQueue_QueueEventParamUnk(stream->event_queue, MEMediaSample, &GUID_NULL, S_OK,
-                (IUnknown *)sample);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        IMFSample_Release(sample);
-    }
+    hr = IMFMediaEventQueue_QueueEventParamUnk(stream->event_queue, MEMediaSample, &GUID_NULL, S_OK,
+            (IUnknown *)sample);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    IMFSample_Release(sample);
 
     return S_OK;
 }
@@ -1770,7 +1473,6 @@
     stream->source = source;
     IMFMediaSource_AddRef(stream->source);
     stream->is_new = TRUE;
-    stream->sample_duration = 333667;
 
     IMFMediaSource_CreatePresentationDescriptor(source, &pd);
     IMFPresentationDescriptor_GetStreamDescriptorByIndex(pd, stream_index, &selected, &stream->sd);
@@ -1976,6 +1678,10 @@
 
     EnterCriticalSection(&source->cs);
 
+    event_type = source->state != SOURCE_STOPPED ? MESourceSeeked : MESourceStarted;
+    hr = IMFMediaEventQueue_QueueEventParamVar(source->event_queue, event_type, &GUID_NULL, S_OK, NULL);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+
     for (i = 0; i < source->stream_count; ++i)
     {
         if (!is_stream_selected(pd, i))
@@ -1988,24 +1694,11 @@
         hr = IMFMediaEventQueue_QueueEventParamVar(source->event_queue, event_type, &GUID_NULL, S_OK, &var);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
-        if (start_position->vt == VT_I8)
-        {
-            source->streams[i]->sample_time = start_position->hVal.QuadPart;
-            event_type = MEStreamSeeked;
-        }
-        else
-        {
-            event_type = MEStreamStarted;
-        }
-
+        event_type = source->state != SOURCE_STOPPED ? MEStreamSeeked : MEStreamStarted;
         hr = IMFMediaEventQueue_QueueEventParamVar(source->streams[i]->event_queue, event_type, &GUID_NULL,
-                S_OK, start_position);
+                S_OK, NULL);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     }
-
-    event_type = start_position->vt == VT_I8 ? MESourceSeeked : MESourceStarted;
-    hr = IMFMediaEventQueue_QueueEventParamVar(source->event_queue, event_type, &GUID_NULL, S_OK, start_position);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
     source->state = SOURCE_RUNNING;
 
@@ -2058,6 +1751,10 @@
 
     EnterCriticalSection(&source->cs);
 
+    event_type = MESourcePaused;
+    hr = IMFMediaEventQueue_QueueEventParamVar(source->event_queue, event_type, &GUID_NULL, S_OK, NULL);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+
     for (i = 0; i < source->stream_count; ++i)
     {
         if (!is_stream_selected(source->pd, i))
@@ -2068,10 +1765,6 @@
                 S_OK, NULL);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     }
-
-    event_type = MESourcePaused;
-    hr = IMFMediaEventQueue_QueueEventParamVar(source->event_queue, event_type, &GUID_NULL, S_OK, NULL);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
     source->state = SOURCE_PAUSED;
     LeaveCriticalSection(&source->cs);
@@ -2388,13 +2081,14 @@
         ATTR_UINT32(MF_MT_AUDIO_BITS_PER_SAMPLE, 2 * 8),
     };
 
+    struct test_stream_sink stream_sink = test_stream_sink;
+    struct test_media_sink media_sink = test_media_sink;
+    struct test_handler handler = test_handler;
     struct test_stub_source *source_impl;
     IMFAsyncCallback *callback, *callback2;
     IMFMediaType *input_type, *output_type;
     IMFTopologyNode *src_node, *sink_node;
-    struct test_media_sink *media_sink;
     IMFPresentationDescriptor *pd;
-    struct test_handler *handler;
     IMFMediaSession *session;
     IMFStreamDescriptor *sd;
     IMFAsyncResult *result;
@@ -2405,8 +2099,8 @@
     HRESULT hr;
     ULONG ref;
 
-    handler = create_test_handler();
-    media_sink = create_test_media_sink(&handler->IMFMediaTypeHandler_iface);
+    stream_sink.handler = &handler.IMFMediaTypeHandler_iface;
+    stream_sink.media_sink = &media_sink.IMFMediaSink_iface;
 
     hr = MFStartup(MF_VERSION, MFSTARTUP_FULL);
     ok(hr == S_OK, "Startup failure, hr %#lx.\n", hr);
@@ -2534,13 +2228,13 @@
 
     hr = MFCreateTopologyNode(MF_TOPOLOGY_OUTPUT_NODE, &sink_node);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    init_sink_node(media_sink->stream, -1, sink_node);
+    init_sink_node(&stream_sink.IMFStreamSink_iface, -1, sink_node);
 
     hr = MFCreateMediaType(&output_type);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     init_media_type(output_type, audio_pcm_48000, -1);
-    handler->media_types_count = 1;
-    handler->media_types = &output_type;
+    handler.media_types_count = 1;
+    handler.media_types = &output_type;
 
     hr = MFCreateTopologyNode(MF_TOPOLOGY_SOURCESTREAM_NODE, &src_node);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2572,13 +2266,15 @@
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(!media_sink->shutdown, "media sink is shutdown.\n");
-    reset_test_media_sink(media_sink);
+    ok(!media_sink.shutdown, "media sink is shutdown.\n");
+    media_sink.shutdown = FALSE;
 
     /* sometimes briefly leaking */
     IMFMediaSession_Release(session);
 
-    test_handler_clear_current_type(handler);
+    if (handler.current_type)
+        IMFMediaType_Release(handler.current_type);
+    handler.current_type = NULL;
 
 
     /* SetTopology without a current output type */
@@ -2596,9 +2292,9 @@
     ok(propvar.punkVal == (IUnknown *)topology, "got punkVal %p\n", propvar.punkVal);
     PropVariantClear(&propvar);
 
-    ok(!handler->enum_count, "got %lu GetMediaTypeByIndex\n", handler->enum_count);
-    ok(!handler->set_current_count, "got %lu SetCurrentMediaType\n", handler->set_current_count);
-    handler->enum_count = handler->set_current_count = 0;
+    ok(!handler.enum_count, "got %lu GetMediaTypeByIndex\n", handler.enum_count);
+    ok(!handler.set_current_count, "got %lu SetCurrentMediaType\n", handler.set_current_count);
+    handler.enum_count = handler.set_current_count = 0;
 
     hr = IMFMediaSession_ClearTopologies(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2608,13 +2304,15 @@
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_sink->shutdown, "media sink didn't shutdown.\n");
-    reset_test_media_sink(media_sink);
+    ok(media_sink.shutdown, "media sink didn't shutdown.\n");
+    media_sink.shutdown = FALSE;
 
     /* sometimes briefly leaking */
     IMFMediaSession_Release(session);
 
-    test_handler_clear_current_type(handler);
+    if (handler.current_type)
+        IMFMediaType_Release(handler.current_type);
+    handler.current_type = NULL;
 
 
     /* SetTopology without a current output type */
@@ -2623,12 +2321,7 @@
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
     hr = IMFMediaSession_SetTopology(session, 0, topology);
-    ok(hr == S_OK || hr == E_INVALIDARG, "Unexpected hr %#lx.\n", hr);
-    if (hr == E_INVALIDARG)
-    {
-        skip("Skipping tests invalid on this windows version.\n");
-        goto skip_invalid;
-    }
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     hr = wait_media_event(session, callback, MESessionTopologySet, 1000, &propvar);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     ok(propvar.vt == VT_UNKNOWN, "got vt %u\n", propvar.vt);
@@ -2636,9 +2329,9 @@
     PropVariantClear(&propvar);
 
     todo_wine
-    ok(!handler->enum_count, "got %lu GetMediaTypeByIndex\n", handler->enum_count);
-    ok(handler->set_current_count, "got %lu SetCurrentMediaType\n", handler->set_current_count);
-    handler->enum_count = handler->set_current_count = 0;
+    ok(!handler.enum_count, "got %lu GetMediaTypeByIndex\n", handler.enum_count);
+    ok(handler.set_current_count, "got %lu SetCurrentMediaType\n", handler.set_current_count);
+    handler.enum_count = handler.set_current_count = 0;
 
     hr = IMFMediaSession_ClearTopologies(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2648,18 +2341,20 @@
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_sink->shutdown, "media sink didn't shutdown.\n");
-    reset_test_media_sink(media_sink);
+    ok(media_sink.shutdown, "media sink didn't shutdown.\n");
+    media_sink.shutdown = FALSE;
 
     /* sometimes briefly leaking */
     IMFMediaSession_Release(session);
 
-    test_handler_clear_current_type(handler);
+    if (handler.current_type)
+        IMFMediaType_Release(handler.current_type);
+    handler.current_type = NULL;
 
 
     /* SetTopology without a current output type, refusing input type */
 
-    handler->invalid_type = input_type;
+    handler.invalid_type = input_type;
 
     hr = MFCreateMediaSession(NULL, &session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2672,9 +2367,9 @@
     ok(propvar.punkVal != (IUnknown *)topology, "got punkVal %p\n", propvar.punkVal);
     PropVariantClear(&propvar);
 
-    ok(handler->enum_count, "got %lu GetMediaTypeByIndex\n", handler->enum_count);
-    ok(handler->set_current_count, "got %lu SetCurrentMediaType\n", handler->set_current_count);
-    handler->enum_count = handler->set_current_count = 0;
+    ok(handler.enum_count, "got %lu GetMediaTypeByIndex\n", handler.enum_count);
+    ok(handler.set_current_count, "got %lu SetCurrentMediaType\n", handler.set_current_count);
+    handler.enum_count = handler.set_current_count = 0;
 
     hr = IMFMediaSession_ClearTopologies(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2684,19 +2379,21 @@
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_sink->shutdown, "media sink didn't shutdown.\n");
-    reset_test_media_sink(media_sink);
+    ok(media_sink.shutdown, "media sink didn't shutdown.\n");
+    media_sink.shutdown = FALSE;
 
     /* sometimes briefly leaking */
     IMFMediaSession_Release(session);
 
-    test_handler_clear_current_type(handler);
+    if (handler.current_type)
+        IMFMediaType_Release(handler.current_type);
+    handler.current_type = NULL;
 
 
     /* SetTopology without a current output type, refusing input type, requiring a converter */
 
-    handler->media_types_count = 0;
-    handler->invalid_type = input_type;
+    handler.media_types_count = 0;
+    handler.invalid_type = input_type;
 
     hr = MFCreateMediaSession(NULL, &session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2709,9 +2406,9 @@
     ok(propvar.punkVal != (IUnknown *)topology, "got punkVal %p\n", propvar.punkVal);
     PropVariantClear(&propvar);
 
-    ok(!handler->enum_count, "got %lu GetMediaTypeByIndex\n", handler->enum_count);
-    ok(handler->set_current_count, "got %lu SetCurrentMediaType\n", handler->set_current_count);
-    handler->enum_count = handler->set_current_count = 0;
+    ok(!handler.enum_count, "got %lu GetMediaTypeByIndex\n", handler.enum_count);
+    ok(handler.set_current_count, "got %lu SetCurrentMediaType\n", handler.set_current_count);
+    handler.enum_count = handler.set_current_count = 0;
 
     hr = IMFMediaSession_ClearTopologies(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2721,19 +2418,21 @@
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_sink->shutdown, "media sink didn't shutdown.\n");
-    reset_test_media_sink(media_sink);
+    ok(media_sink.shutdown, "media sink didn't shutdown.\n");
+    media_sink.shutdown = FALSE;
 
     /* sometimes briefly leaking */
     IMFMediaSession_Release(session);
 
-    test_handler_clear_current_type(handler);
+    if (handler.current_type)
+        IMFMediaType_Release(handler.current_type);
+    handler.current_type = NULL;
 
 
     /* SetTopology with a current output type */
 
-    handler->media_types_count = 1;
-    IMFMediaType_AddRef((handler->current_type = output_type));
+    handler.media_types_count = 1;
+    IMFMediaType_AddRef((handler.current_type = output_type));
 
     hr = MFCreateMediaSession(NULL, &session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2746,9 +2445,9 @@
     ok(propvar.punkVal != (IUnknown *)topology, "got punkVal %p\n", propvar.punkVal);
     PropVariantClear(&propvar);
 
-    ok(!handler->enum_count, "got %lu GetMediaTypeByIndex\n", handler->enum_count);
-    ok(handler->set_current_count, "got %lu SetCurrentMediaType\n", handler->set_current_count);
-    handler->enum_count = handler->set_current_count = 0;
+    ok(!handler.enum_count, "got %lu GetMediaTypeByIndex\n", handler.enum_count);
+    ok(handler.set_current_count, "got %lu SetCurrentMediaType\n", handler.set_current_count);
+    handler.enum_count = handler.set_current_count = 0;
 
     hr = IMFMediaSession_ClearTopologies(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2758,8 +2457,8 @@
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_sink->shutdown, "media sink didn't shutdown.\n");
-    reset_test_media_sink(media_sink);
+    ok(media_sink.shutdown, "media sink didn't shutdown.\n");
+    media_sink.shutdown = FALSE;
 
     /* sometimes briefly leaking */
     IMFMediaSession_Release(session);
@@ -2802,8 +2501,8 @@
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_sink->shutdown, "media sink didn't shutdown.\n");
-    reset_test_media_sink(media_sink);
+    ok(media_sink.shutdown, "media sink didn't shutdown.\n");
+    media_sink.shutdown = FALSE;
 
     source_impl->begin_get_event_res = E_NOTIMPL;
 
@@ -2851,7 +2550,8 @@
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_sink->shutdown, "media sink didn't shutdown.\n");
+    ok(media_sink.shutdown, "media sink didn't shutdown.\n");
+    media_sink.shutdown = FALSE;
 
     source_impl->begin_get_event_res = E_NOTIMPL;
 
@@ -2859,14 +2559,14 @@
     CLEAR_CALLED(test_stub_source_QueueEvent);
     CLEAR_CALLED(test_stub_source_Start);
 
-skip_invalid:
-    IMFMediaTypeHandler_Release(&handler->IMFMediaTypeHandler_iface);
-    IMFMediaSink_Release(&media_sink->IMFMediaSink_iface);
-
     /* sometimes briefly leaking */
     IMFMediaSession_Release(session);
 
     IMFAsyncCallback_Release(callback);
+
+    if (handler.current_type)
+        IMFMediaType_Release(handler.current_type);
+    handler.current_type = NULL;
 
     hr = IMFTopology_Clear(topology);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -2911,8 +2611,6 @@
     hr = MFCreateMediaSession(NULL, &session);
     ok(hr == S_OK, "Failed to create media session, hr %#lx.\n", hr);
 
-    check_service_interface(session, &InvalidServiceGUID, &InvalidServiceGUID, FALSE);
-
     check_interface(session, &IID_IMFGetService, TRUE);
     check_interface(session, &IID_IMFRateSupport, TRUE);
     check_interface(session, &IID_IMFRateControl, TRUE);
@@ -3197,12 +2895,6 @@
     IMFSample *sample;
     HRESULT hr;
     DWORD res;
-
-    if (grabber->ready_event && !grabber->done_event)
-    {
-        SetEvent(grabber->ready_event);
-        return S_OK;
-    }
 
     if (!grabber->ready_event)
         return E_NOTIMPL;
@@ -4305,10 +3997,10 @@
     IMFSampleGrabberSinkCallback_Release(grabber_callback);
 }
 
-/* create a test topology with the specified source, sink, and option MFT. Return duration if required */
-static IMFTopology *create_test_topology_unk(IMFMediaSource *source, IUnknown *sink, IUnknown *mft, UINT64 *duration)
-{
-    IMFTopologyNode *src_node, *sink_node, *mft_node;
+/* create a test topology with the specified source and sink, and return duration if required */
+static IMFTopology *create_test_topology(IMFMediaSource *source, IMFActivate *sink_activate, UINT64 *duration)
+{
+    IMFTopologyNode *src_node, *sink_node;
     IMFPresentationDescriptor *pd;
     IMFTopology *topology = NULL;
     IMFStreamDescriptor *sd;
@@ -4325,24 +4017,8 @@
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     hr = IMFTopology_AddNode(topology, src_node);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    if (mft)
-    {
-        hr = MFCreateTopologyNode(MF_TOPOLOGY_TRANSFORM_NODE, &mft_node);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        hr = IMFTopology_AddNode(topology, mft_node);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        hr = IMFTopologyNode_ConnectOutput(src_node, 0, mft_node, 0);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        hr = IMFTopologyNode_ConnectOutput(mft_node, 0, sink_node, 0);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        hr = IMFTopologyNode_SetObject(mft_node, mft);
-        ok(hr == S_OK, "Failed to set object, hr %#lx.\n", hr);
-    }
-    else
-    {
-        hr = IMFTopologyNode_ConnectOutput(src_node, 0, sink_node, 0);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    }
+    hr = IMFTopologyNode_ConnectOutput(src_node, 0, sink_node, 0);
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     hr = IMFMediaSource_CreatePresentationDescriptor(source, &pd);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     hr = IMFPresentationDescriptor_GetStreamDescriptorByIndex(pd, 0, &selected, &sd);
@@ -4354,7 +4030,7 @@
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     }
     init_source_node(source, -1, src_node, pd, sd);
-    hr = IMFTopologyNode_SetObject(sink_node, sink);
+    hr = IMFTopologyNode_SetObject(sink_node, (IUnknown *)sink_activate);
     ok(hr == S_OK, "Failed to set object, hr %#lx.\n", hr);
     hr = IMFTopologyNode_SetUINT32(sink_node, &MF_TOPONODE_CONNECT_METHOD, MF_CONNECT_ALLOW_DECODER);
     ok(hr == S_OK, "Failed to set connect method, hr %#lx.\n", hr);
@@ -4366,11 +4042,6 @@
     IMFTopologyNode_Release(src_node);
     IMFTopologyNode_Release(sink_node);
     return topology;
-}
-
-static IMFTopology *create_test_topology(IMFMediaSource *source, IMFActivate *sink_activate, UINT64 *duration)
-{
-    return create_test_topology_unk(source, (IUnknown*)sink_activate, NULL, duration);
 }
 
 static void test_sample_grabber_orientation(GUID subtype)
@@ -4736,27 +4407,21 @@
 
     IMFPresentationClock *present_clock, *present_clock2;
     IMFMediaType *mediatype, *mediatype2, *mediatype3;
-    UINT32 channel_count, rate, bytes_per_second;
     IMFClockStateSink *state_sink, *state_sink2;
     IMFMediaTypeHandler *handler, *handler2;
     IMFPresentationTimeSource *time_source;
     IMFSimpleAudioVolume *simple_volume;
     IMFAudioStreamVolume *stream_volume;
-    IMFAsyncCallback *callback;
     IMFMediaSink *sink, *sink2;
     IMFStreamSink *stream_sink;
+    UINT32 channel_count, rate;
     IMFAttributes *attributes;
-    IMFMediaBuffer *buffer;
     DWORD id, flags, count;
     IMFActivate *activate;
-    IMFMediaEvent *event;
     MFCLOCK_STATE state;
-    PROPVARIANT propvar;
-    IMFSample *sample;
     IMFClock *clock;
     IUnknown *unk;
     HRESULT hr;
-    BYTE *buff;
     GUID guid;
     BOOL mute;
     LONG ref;
@@ -4978,7 +4643,6 @@
     hr = IMFMediaTypeHandler_GetCurrentMediaType(handler, &mediatype);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     ok(mediatype == mediatype2, "Unexpected instance.\n");
-    IMFMediaType_GetUINT32(mediatype, &MF_MT_AUDIO_AVG_BYTES_PER_SECOND, &bytes_per_second);
     IMFMediaType_Release(mediatype);
 
     IMFMediaType_Release(mediatype2);
@@ -5019,117 +4683,6 @@
 
     hr = IMFClockStateSink_OnClockStop(state_sink, 0);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    callback = create_test_callback(TRUE);
-
-    /* Flush events */
-    while (SUCCEEDED(IMFStreamSink_GetEvent(stream_sink, MF_EVENT_FLAG_NO_WAIT, &event)))
-        IMFMediaEvent_Release(event);
-
-    hr = IMFClockStateSink_OnClockStart(state_sink, 0, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    propvar.vt = VT_EMPTY;
-    hr = gen_wait_media_event((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkStarted, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = gen_wait_media_event((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkRequestSample, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = MFCreateMemoryBuffer(bytes_per_second, &buffer);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFMediaBuffer_Lock(buffer, &buff, NULL, NULL);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    memset(buff, 0, bytes_per_second);
-    hr = IMFMediaBuffer_Unlock(buffer);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFMediaBuffer_SetCurrentLength(buffer, bytes_per_second);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = MFCreateSample(&sample);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFSample_AddBuffer(sample, buffer);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFSample_SetSampleTime(sample, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFStreamSink_ProcessSample(stream_sink, sample);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    IMFSample_Release(sample);
-
-    hr = gen_wait_media_event((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkRequestSample, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = MFCreateSample(&sample);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFSample_AddBuffer(sample, buffer);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    IMFMediaBuffer_Release(buffer);
-
-    hr = IMFSample_SetSampleTime(sample, 10000000);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFStreamSink_ProcessSample(stream_sink, sample);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    IMFSample_Release(sample);
-
-    hr = IMFClockStateSink_OnClockPause(state_sink, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = gen_wait_media_event((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkPaused, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFStreamSink_Flush(stream_sink);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    /* confirm no new sample is requested after a flush */
-    hr = gen_wait_media_event_until_blocking((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkRequestSample, 1000, &propvar);
-    todo_wine
-    ok(hr == WAIT_TIMEOUT, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFClockStateSink_OnClockStart(state_sink, 0, 123456);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = gen_wait_media_event_until_blocking((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkStarted, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    /* they are only requested after a call to OnClockStart */
-    hr = gen_wait_media_event_until_blocking((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkRequestSample, 1000, &propvar);
-    todo_wine
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = gen_wait_media_event_until_blocking((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkRequestSample, 1000, &propvar);
-    todo_wine
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    /* but if the original requests aren't satisfied ... */
-    hr = IMFClockStateSink_OnClockPause(state_sink, 0);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = gen_wait_media_event((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkPaused, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFStreamSink_Flush(stream_sink);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    /* ... there is still no new sample request after a flush ... */
-    hr = gen_wait_media_event_until_blocking((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkRequestSample, 1000, &propvar);
-    ok(hr == WAIT_TIMEOUT, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFClockStateSink_OnClockStart(state_sink, 0, 654321);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = gen_wait_media_event_until_blocking((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkStarted, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    /* ... and still none after a call to OnClockStart. The client must keep track of these pending requests. */
-    hr = gen_wait_media_event_until_blocking((IMFMediaEventGenerator*)stream_sink, callback, MEStreamSinkRequestSample, 1000, &propvar);
-    ok(hr == WAIT_TIMEOUT, "Unexpected hr %#lx.\n", hr);
-
-    IMFAsyncCallback_Release(callback);
 
     IMFClockStateSink_Release(state_sink);
 
@@ -5993,8 +5546,6 @@
         L"http://test.winehq.org",
     };
 
-    INetworkListManager *network_list_manager;
-    NLM_CONNECTIVITY connectivity;
     IMFSourceResolver *resolver;
     IMFByteStream *byte_stream;
     IMFAttributes *attributes;
@@ -6004,20 +5555,6 @@
     UINT64 length;
     DWORD i, caps;
     HRESULT hr;
-
-    hr = CoInitialize(NULL);
-    ok(hr == S_OK, "Failed to initialize, hr %#lx\n", hr);
-    hr = CoCreateInstance(&CLSID_NetworkListManager, NULL, CLSCTX_INPROC_SERVER, &IID_INetworkListManager, (void **)&network_list_manager);
-    ok(hr == S_OK, "got hr %#lx\n", hr);
-    hr = INetworkListManager_GetConnectivity(network_list_manager, &connectivity);
-    ok(hr == S_OK, "got hr %#lx\n", hr);
-    INetworkListManager_Release(network_list_manager);
-    CoUninitialize();
-    if (connectivity == NLM_CONNECTIVITY_DISCONNECTED)
-    {
-        skip("Internet connection unavailable, skipping scheme resolver tests.\n");
-        return;
-    }
 
     hr = MFStartup(MF_VERSION, MFSTARTUP_FULL);
     ok(hr == S_OK, "got hr %#lx\n", hr);
@@ -6109,7 +5646,7 @@
     ok(hr == NS_E_FILE_NOT_FOUND, "got hr %#lx\n", hr);
     hr = IMFSourceResolver_CreateObjectFromURL(resolver, L"mms://test.winehq.org/tests/invalid.mp3", MF_RESOLUTION_BYTESTREAM, NULL, &type, &object);
     todo_wine
-    ok(hr == MF_E_UNSUPPORTED_BYTESTREAM_TYPE || hr == NS_E_FILE_NOT_FOUND, "got hr %#lx\n", hr);
+    ok(hr == MF_E_UNSUPPORTED_BYTESTREAM_TYPE, "got hr %#lx\n", hr);
 
     IMFSourceResolver_Release(resolver);
 
@@ -6128,7 +5665,7 @@
     };
     IMFMediaType *media_type, *input_types[2], *output_types[2];
     IMFStreamDescriptor *input_descriptor, *output_descriptor;
-    struct test_stream_sink *stream_sink;
+    struct test_stream_sink stream_sink = test_stream_sink;
     IMFMediaTypeHandler *input_handler, *output_handler;
     IMFTransform *transform;
     IMFTopologyNode *node;
@@ -6283,8 +5820,8 @@
     hr = pMFGetTopoNodeCurrentType(node, 0, TRUE, &media_type);
     ok(hr == MF_E_INVALIDSTREAMNUMBER, "Unexpected hr %#lx.\n", hr);
 
-    stream_sink = create_test_stream_sink(NULL, output_handler, FALSE);
-    hr = IMFTopologyNode_SetObject(node, (IUnknown *)&stream_sink->IMFStreamSink_iface);
+    stream_sink.handler = output_handler;
+    hr = IMFTopologyNode_SetObject(node, (IUnknown *)&stream_sink.IMFStreamSink_iface);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     hr = pMFGetTopoNodeCurrentType(node, 0, FALSE, &media_type);
     ok(hr == MF_E_NOT_INITIALIZED, "Unexpected hr %#lx.\n", hr);
@@ -6298,7 +5835,6 @@
 
     ref = IMFTopologyNode_Release(node);
     ok(ref == 0, "Release returned %ld\n", ref);
-    IMFStreamSink_Release(&stream_sink->IMFStreamSink_iface);
 
 
     /* Transform node. */
@@ -6444,8 +5980,8 @@
 {
     IMFMediaSink *sink = NULL, *sink2 = NULL, *sink_audio = NULL, *sink_video = NULL, *sink_empty = NULL;
     IMFByteStream *bytestream, *bytestream_audio, *bytestream_video, *bytestream_empty;
-    IMFMediaType *audio_type, *video_type, *media_type, *media_type_out;
     DWORD id, count, flags, width = 96, height = 96;
+    IMFMediaType *audio_type, *video_type, *media_type;
     IMFMediaTypeHandler *type_handler = NULL;
     IMFPresentationClock *clock;
     IMFStreamSink *stream_sink;
@@ -6669,27 +6205,22 @@
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
     hr = IMFMediaTypeHandler_GetMajorType(type_handler, NULL);
+    todo_wine
     ok(hr == E_POINTER, "Unexpected hr %#lx.\n", hr);
     hr = IMFMediaTypeHandler_GetMajorType(type_handler, &guid);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    todo_wine
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    todo_wine
     ok(IsEqualGUID(&guid, &MFMediaType_Audio), "Unexpected major type.\n");
 
-    hr = IMFMediaTypeHandler_GetMediaTypeCount(type_handler, NULL);
-    ok(hr == E_POINTER, "Unexpected hr %#lx.\n", hr);
     hr = IMFMediaTypeHandler_GetMediaTypeCount(type_handler, &count);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    todo_wine
+    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
+    todo_wine
     ok(count == 1, "Unexpected count %lu.\n", count);
 
     hr = IMFMediaTypeHandler_GetCurrentMediaType(type_handler, &media_type);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFMediaTypeHandler_GetMediaTypeByIndex(type_handler, 1, &media_type_out);
-    ok(hr == MF_E_NO_MORE_TYPES, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaTypeHandler_GetMediaTypeByIndex(type_handler, 0, NULL);
-    ok(hr == E_POINTER, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaTypeHandler_GetMediaTypeByIndex(type_handler, 0, &media_type_out);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_type_out == media_type, "Got different media type pointer.\n");
 
     hr = IMFMediaType_SetUINT32(media_type, &MF_MT_AUDIO_NUM_CHANNELS, 1);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -6717,8 +6248,10 @@
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
     hr = IMFMediaTypeHandler_GetMajorType(type_handler, NULL);
+    todo_wine
     ok(hr == E_POINTER, "Unexpected hr %#lx.\n", hr);
     hr = IMFMediaTypeHandler_GetMajorType(type_handler, &guid);
+    todo_wine
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
     IMFStreamSink_Release(stream_sink);
@@ -6764,203 +6297,6 @@
     ok(propvar.vt == VT_UNKNOWN, "Unexpected type %d.\n", propvar.vt);
     ok(!!propvar.punkVal, "Unexpected pointer.\n");
     PropVariantClear(&propvar);
-}
-
-static IMFTransform *topology_get_transform(IMFTopology *topology)
-{
-    IMFTransform *transform;
-    IMFTopologyNode *node;
-    MF_TOPOLOGY_TYPE type;
-    HRESULT hr;
-    WORD count;
-    UINT i;
-
-    hr = IMFTopology_GetNodeCount(topology, &count);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    for (i = 0; i < count; ++i)
-    {
-        hr = IMFTopology_GetNode(topology, i, &node);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        hr = IMFTopologyNode_GetNodeType(node, &type);
-        ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        if (type == MF_TOPOLOGY_TRANSFORM_NODE)
-        {
-            hr = IMFTopologyNode_GetObject(node, (IUnknown **)&transform);
-            ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-            IMFTopologyNode_Release(node);
-            return transform;
-        }
-        IMFTopologyNode_Release(node);
-    }
-
-    return NULL;
-}
-
-#define get_current_media_type_frame_size(a) get_current_media_type_frame_size_(__LINE__, (IUnknown *)a)
-static UINT64 get_current_media_type_frame_size_(int line, IUnknown *unknown)
-{
-    IMFMediaTypeHandler *handler;
-    IMFMediaType *output_type;
-    IMFTransform *transform;
-    UINT64 frame_size;
-    HRESULT hr;
-
-    /* Wine does not currently use a transform. */
-    if (!unknown)
-        return 0;
-
-    if (SUCCEEDED(IUnknown_QueryInterface(unknown, &IID_IMFTransform, (void **)&transform)))
-    {
-        hr = IMFTransform_GetOutputCurrentType(transform, 0, &output_type);
-        ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        hr = IMFMediaType_GetUINT64(output_type, &MF_MT_FRAME_SIZE, &frame_size);
-        ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-        IMFMediaType_Release(output_type);
-        IMFTransform_Release(transform);
-        return frame_size;
-    }
-
-    hr = IUnknown_QueryInterface(unknown, &IID_IMFMediaTypeHandler, (void **)&handler);
-    ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaTypeHandler_GetCurrentMediaType(handler, &output_type);
-    ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaType_GetUINT64(output_type, &MF_MT_FRAME_SIZE, &frame_size);
-    ok_(__FILE__, line)(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    IMFMediaType_Release(output_type);
-    IMFMediaTypeHandler_Release(handler);
-    return frame_size;
-}
-
-static void test_h264_output_alignment(void)
-{
-    media_type_desc video_nv12_desc =
-    {
-        ATTR_GUID(MF_MT_MAJOR_TYPE, MFMediaType_Video),
-        ATTR_GUID(MF_MT_SUBTYPE, MFVideoFormat_NV12),
-    };
-    struct test_grabber_callback *grabber_callback;
-    IMFTopology *topology, *resolved_topology;
-    struct test_callback *test_callback;
-    IMFMediaTypeHandler *handler;
-    IMFActivate *sink_activate;
-    IMFTopoLoader *topo_loader;
-    IMFStreamSink *stream_sink;
-    IMFAsyncCallback *callback;
-    IMFMediaType *output_type;
-    IMFMediaSession *session;
-    IMFMediaSink *media_sink;
-    IMFTransform *transform;
-    IMFMediaSource *source;
-    PROPVARIANT propvar;
-    UINT64 frame_size;
-    UINT32 status;
-    HRESULT hr;
-    DWORD ret;
-
-    hr = MFStartup(MF_VERSION, MFSTARTUP_FULL);
-    ok(hr == S_OK, "Failed to start up, hr %#lx.\n", hr);
-
-    if (!(source = create_media_source(L"test-unaligned.mp4", L"video/mp4")))
-    {
-        todo_wine /* Gitlab CI Debian runner */
-        win_skip("MP4 media source is not supported, skipping tests.\n");
-        MFShutdown();
-        return;
-    }
-
-    grabber_callback = impl_from_IMFSampleGrabberSinkCallback(create_test_grabber_callback());
-    grabber_callback->ready_event = CreateEventW(NULL, FALSE, FALSE, NULL);
-    ok(!!grabber_callback->ready_event, "CreateEventW failed, error %lu\n", GetLastError());
-
-    hr = MFCreateMediaType(&output_type);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    init_media_type(output_type, video_nv12_desc, -1);
-    hr = MFCreateSampleGrabberSinkActivate(output_type, &grabber_callback->IMFSampleGrabberSinkCallback_iface, &sink_activate);
-    ok(hr == S_OK, "Failed to create grabber sink, hr %#lx.\n", hr);
-    IMFMediaType_Release(output_type);
-
-    IMFActivate_ActivateObject(sink_activate, &IID_IMFMediaSink, (void **)&media_sink);
-    hr = IMFMediaSink_GetStreamSinkByIndex(media_sink, 0, &stream_sink);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    IMFActivate_Release(sink_activate);
-    topology = create_test_topology_unk(source, (IUnknown *)stream_sink, NULL, NULL);
-    hr = MFCreateTopoLoader(&topo_loader);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFTopoLoader_Load(topo_loader, topology, &resolved_topology, NULL);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    IMFTopology_Release(topology);
-    IMFTopoLoader_Release(topo_loader);
-
-    hr = IMFStreamSink_GetMediaTypeHandler(stream_sink, &handler);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaTypeHandler_GetCurrentMediaType(handler, &output_type);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaType_GetUINT64(output_type, &MF_MT_FRAME_SIZE, &frame_size);
-    ok(hr == MF_E_ATTRIBUTENOTFOUND, "Unexpected hr %#lx.\n", hr);
-    IMFMediaType_Release(output_type);
-
-    hr = MFCreateMediaSession(NULL, &session);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaSession_SetTopology(session, 0, resolved_topology);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    transform = topology_get_transform(resolved_topology);
-    IMFTopology_Release(resolved_topology);
-
-    callback = create_test_callback(TRUE);
-    hr = wait_media_event(session, callback, MESessionTopologySet, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    test_callback = impl_from_IMFAsyncCallback(callback);
-    hr = wait_media_event(session, callback, MESessionTopologyStatus, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaEvent_GetUINT32(test_callback->media_event, &MF_EVENT_TOPOLOGY_STATUS, &status);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(status == MF_TOPOSTATUS_READY, "Unexpected status %d.\n", status);
-
-    frame_size = get_current_media_type_frame_size(transform);
-    todo_wine
-    ok(frame_size == (((UINT64)64 << 32) | 72), "Unexpected frame size %#llx\n", frame_size);
-    frame_size = get_current_media_type_frame_size(handler);
-    ok(frame_size == (((UINT64)64 << 32) | 72), "Unexpected frame size %#llx\n", frame_size);
-
-    propvar.vt = VT_EMPTY;
-    hr = IMFMediaSession_Start(session, &GUID_NULL, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    /* frame size change occurs before the first sample is delivered */
-    ret = WaitForSingleObject(grabber_callback->ready_event, 1000);
-    ok(ret == WAIT_OBJECT_0, "WaitForSingleObject returned %lu\n", ret);
-    frame_size = get_current_media_type_frame_size(transform);
-    todo_wine
-    ok(frame_size == (((UINT64)64 << 32) | 80), "Unexpected frame size %#llx\n", frame_size);
-    frame_size = get_current_media_type_frame_size(handler);
-    todo_wine
-    ok(frame_size == (((UINT64)64 << 32) | 80), "Unexpected frame size %#llx\n", frame_size);
-
-    hr = IMFMediaSession_Stop(session);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaSession_Close(session);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = wait_media_event(session, callback, MESessionClosed, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = IMFMediaSource_Shutdown(source);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaSession_Shutdown(session);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    IMFMediaTypeHandler_Release(handler);
-    if (transform)
-        IMFTransform_Release(transform);
-    IMFAsyncCallback_Release(callback);
-    IMFMediaSession_Release(session);
-    IMFStreamSink_Release(stream_sink);
-    IMFMediaSink_Release(media_sink);
-    IMFSampleGrabberSinkCallback_Release(&grabber_callback->IMFSampleGrabberSinkCallback_iface);
-    IMFMediaSource_Release(source);
-
-    hr = MFShutdown();
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 }
 
 static void test_media_session_Start(void)
@@ -7302,6 +6638,7 @@
     HRESULT hr;
     enum
     {
+        TEST_TOPOLOGY,
         TEST_START,
         TEST_RESTART,
         TEST_PAUSE,
@@ -7314,7 +6651,7 @@
 
     /* These tests don't cover asynchronous shutdown, which is difficult to consistently test. */
 
-    for (shutdown_point = TEST_START; shutdown_point <= TEST_CLOSE; ++shutdown_point)
+    for (shutdown_point = TEST_TOPOLOGY; shutdown_point <= TEST_CLOSE; ++shutdown_point)
     {
         winetest_push_context("Test %d", shutdown_point);
 
@@ -7338,6 +6675,8 @@
         hr = MFCreateMediaSession(NULL, &session);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
         topology = create_test_topology(source, sink_activate, &duration);
+        if (shutdown_point == TEST_TOPOLOGY)
+            IMFMediaSource_Shutdown(source);
         hr = IMFMediaSession_SetTopology(session, 0, topology);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
         IMFTopology_Release(topology);
@@ -7350,7 +6689,7 @@
             IMFMediaSource_Shutdown(source);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
         hr = wait_media_event(session, callback, MESessionStarted, 5000, &propvar);
-        ok(hr == (shutdown_point == TEST_START ? MF_E_INVALIDREQUEST : S_OK), "Unexpected hr %#lx.\n", hr);
+        ok(hr == (shutdown_point <= TEST_START ? MF_E_INVALIDREQUEST : S_OK), "Unexpected hr %#lx.\n", hr);
 
         switch (shutdown_point)
         {
@@ -7371,32 +6710,30 @@
                 hr = IMFMediaSession_Pause(session);
                 IMFMediaSource_Shutdown(source);
                 ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-                hr = wait_media_event_until_blocking(session, callback, MESessionPaused, 1000, &propvar);
+                hr = wait_media_event(session, callback, MESessionPaused, 1000, &propvar);
                 /* Windows has not been observed to emit PAUSEWHILESTOPPED here, but this could
-                 * be a matter of async command timing, and this error is not exactly wrong.
-                 * Windows occasionally never sends MESessionPaused here; ditto for Stopped and Closed. */
-                ok(hr == S_OK || hr == MF_E_SESSION_PAUSEWHILESTOPPED || hr == WAIT_TIMEOUT || hr == MF_E_SHUTDOWN,
+                 * be a matter of async command timing, and this error is not exactly wrong. */
+                ok(hr == S_OK || hr == MF_E_SESSION_PAUSEWHILESTOPPED || hr == MF_E_SHUTDOWN,
                         "Unexpected hr %#lx.\n", hr);
                 break;
             case TEST_STOP:
                 hr = IMFMediaSession_Stop(session);
                 IMFMediaSource_Shutdown(source);
                 ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-                hr = wait_media_event_until_blocking(session, callback, MESessionStopped, 1000, &propvar);
-                ok(hr == S_OK || hr == WAIT_TIMEOUT || hr == MF_E_SHUTDOWN, "Unexpected hr %#lx.\n", hr);
+                hr = wait_media_event(session, callback, MESessionStopped, 1000, &propvar);
+                ok(hr == S_OK || hr == MF_E_SHUTDOWN, "Unexpected hr %#lx.\n", hr);
                 break;
             case TEST_CLOSE:
                 hr = IMFMediaSession_Close(session);
                 IMFMediaSource_Shutdown(source);
                 ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-                hr = wait_media_event_until_blocking(session, callback, MESessionClosed, 1000, &propvar);
-                ok(hr == S_OK || hr == WAIT_TIMEOUT || hr == MF_E_SHUTDOWN, "Unexpected hr %#lx.\n", hr);
+                hr = wait_media_event(session, callback, MESessionClosed, 1000, &propvar);
+                ok(hr == S_OK || hr == MF_E_SHUTDOWN, "Unexpected hr %#lx.\n", hr);
             default:
                 break;
         }
 
-        /* Skip tests where the results in Windows are too uncertain to be worth checking. */
-        if (shutdown_point >= TEST_PAUSE)
+        if (shutdown_point == TEST_CLOSE)
             goto done;
 
         IMFMediaSource_Release(source);
@@ -7438,7 +6775,10 @@
         hr = IMFMediaSession_Close(session);
         ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
         hr = wait_media_event_until_blocking(session, callback, MESessionClosed, 1000, &propvar);
-        ok(hr == MF_E_SHUTDOWN || hr == S_OK, "Unexpected hr %#lx.\n", hr);
+        if (shutdown_point >= TEST_PAUSE)
+            ok(hr == MF_E_SHUTDOWN || hr == S_OK || hr == WAIT_TIMEOUT, "Unexpected hr %#lx.\n", hr);
+        else
+            ok(hr == MF_E_SHUTDOWN || hr == S_OK, "Unexpected hr %#lx.\n", hr);
 
 done:
         hr = IMFMediaSession_Shutdown(session);
@@ -7630,7 +6970,7 @@
     hr = IMFMediaSource_Shutdown(source);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
     hr = wait_media_event_until_blocking(session, callback, MESessionClosed, 5000, &propvar);
-    ok(hr == MF_E_SHUTDOWN || hr == WAIT_TIMEOUT, "Unexpected hr %#lx.\n", hr);
+    ok(hr == MF_E_SHUTDOWN, "Unexpected hr %#lx.\n", hr);
 
     hr = IMFMediaSession_Shutdown(session);
     ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
@@ -7787,653 +7127,6 @@
 
     hr = MFShutdown();
     ok(hr == S_OK, "Shutdown failure, hr %#lx.\n", hr);
-}
-
-struct test_transform
-{
-    IMFTransform IMFTransform_iface;
-    LONG refcount;
-
-    const MFT_OUTPUT_STREAM_INFO *output_stream_info;
-
-    UINT input_count;
-    IMFMediaType **input_types;
-    IMFMediaType *input_type;
-
-    UINT output_count;
-    IMFMediaType **output_types;
-    IMFMediaType *output_type;
-
-    IMFSample *output;
-};
-
-static struct test_transform *test_transform_from_IMFTransform(IMFTransform *iface)
-{
-    return CONTAINING_RECORD(iface, struct test_transform, IMFTransform_iface);
-}
-
-static HRESULT WINAPI test_transform_QueryInterface(IMFTransform *iface, REFIID iid, void **out)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-
-    if (IsEqualGUID(iid, &IID_IUnknown)
-            || IsEqualGUID(iid, &IID_IMFTransform))
-    {
-        IMFTransform_AddRef(&transform->IMFTransform_iface);
-        *out = &transform->IMFTransform_iface;
-        return S_OK;
-    }
-
-    *out = NULL;
-    return E_NOINTERFACE;
-}
-
-static ULONG WINAPI test_transform_AddRef(IMFTransform *iface)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    ULONG refcount = InterlockedIncrement(&transform->refcount);
-    return refcount;
-}
-
-static ULONG WINAPI test_transform_Release(IMFTransform *iface)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    ULONG refcount = InterlockedDecrement(&transform->refcount);
-
-    if (!refcount)
-    {
-        if (transform->input_type)
-            IMFMediaType_Release(transform->input_type);
-        if (transform->output_type)
-            IMFMediaType_Release(transform->output_type);
-        free(transform);
-    }
-
-    return refcount;
-}
-
-static HRESULT WINAPI test_transform_GetStreamLimits(IMFTransform *iface, DWORD *input_minimum,
-        DWORD *input_maximum, DWORD *output_minimum, DWORD *output_maximum)
-{
-    ok(0, "Unexpected call.\n");
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_GetStreamCount(IMFTransform *iface, DWORD *inputs, DWORD *outputs)
-{
-    *inputs = *outputs = 1;
-    return S_OK;
-}
-
-static HRESULT WINAPI test_transform_GetStreamIDs(IMFTransform *iface, DWORD input_size, DWORD *inputs,
-        DWORD output_size, DWORD *outputs)
-{
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_GetInputStreamInfo(IMFTransform *iface, DWORD id, MFT_INPUT_STREAM_INFO *info)
-{
-    ok(0, "Unexpected %s call.\n", __func__);
-    return E_NOTIMPL;
-}
-
-static void test_transform_set_output_stream_info(IMFTransform *iface, const MFT_OUTPUT_STREAM_INFO *info)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    transform->output_stream_info = info;
-}
-
-static HRESULT WINAPI test_transform_GetOutputStreamInfo(IMFTransform *iface, DWORD id, MFT_OUTPUT_STREAM_INFO *info)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-
-    ok(!!transform->output_stream_info, "Unexpected %s iface %p call.\n", __func__, iface);
-    if (!transform->output_stream_info)
-        return E_NOTIMPL;
-
-    *info = *transform->output_stream_info;
-    return S_OK;
-}
-
-static HRESULT WINAPI test_transform_GetAttributes(IMFTransform *iface, IMFAttributes **attributes)
-{
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_GetInputStreamAttributes(IMFTransform *iface, DWORD id, IMFAttributes **attributes)
-{
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_GetOutputStreamAttributes(IMFTransform *iface, DWORD id, IMFAttributes **attributes)
-{
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_DeleteInputStream(IMFTransform *iface, DWORD id)
-{
-    ok(0, "Unexpected %s call.\n", __func__);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_AddInputStreams(IMFTransform *iface, DWORD streams, DWORD *ids)
-{
-    ok(0, "Unexpected %s call.\n", __func__);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_GetInputAvailableType(IMFTransform *iface, DWORD id, DWORD index,
-        IMFMediaType **type)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-
-    if (index >= transform->input_count)
-    {
-        *type = NULL;
-        return MF_E_NO_MORE_TYPES;
-    }
-
-    *type = transform->input_types[index];
-    IMFMediaType_AddRef(*type);
-    return S_OK;
-}
-
-static HRESULT WINAPI test_transform_GetOutputAvailableType(IMFTransform *iface, DWORD id,
-        DWORD index, IMFMediaType **type)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-
-    if (index >= transform->output_count)
-    {
-        *type = NULL;
-        return MF_E_NO_MORE_TYPES;
-    }
-
-    *type = transform->output_types[index];
-    IMFMediaType_AddRef(*type);
-    return S_OK;
-}
-
-static HRESULT WINAPI test_transform_SetInputType(IMFTransform *iface, DWORD id, IMFMediaType *type, DWORD flags)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    if (flags & MFT_SET_TYPE_TEST_ONLY)
-        return S_OK;
-    if (transform->input_type)
-        IMFMediaType_Release(transform->input_type);
-    if ((transform->input_type = type))
-        IMFMediaType_AddRef(transform->input_type);
-    return S_OK;
-}
-
-static HRESULT WINAPI test_transform_SetOutputType(IMFTransform *iface, DWORD id, IMFMediaType *type, DWORD flags)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    if (flags & MFT_SET_TYPE_TEST_ONLY)
-        return S_OK;
-    if (transform->output_type)
-        IMFMediaType_Release(transform->output_type);
-    if ((transform->output_type = type))
-        IMFMediaType_AddRef(transform->output_type);
-    return S_OK;
-}
-
-static HRESULT WINAPI test_transform_GetInputCurrentType(IMFTransform *iface, DWORD id, IMFMediaType **type)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    if (!(*type = transform->input_type))
-        return MF_E_TRANSFORM_TYPE_NOT_SET;
-    IMFMediaType_AddRef(*type);
-    return S_OK;
-}
-
-static HRESULT WINAPI test_transform_GetOutputCurrentType(IMFTransform *iface, DWORD id, IMFMediaType **type)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    if (!(*type = transform->output_type))
-        return MF_E_TRANSFORM_TYPE_NOT_SET;
-    IMFMediaType_AddRef(*type);
-    return S_OK;
-}
-
-static HRESULT WINAPI test_transform_GetInputStatus(IMFTransform *iface, DWORD id, DWORD *flags)
-{
-    ok(0, "Unexpected %s call.\n", __func__);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_GetOutputStatus(IMFTransform *iface, DWORD *flags)
-{
-    ok(0, "Unexpected %s call.\n", __func__);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_SetOutputBounds(IMFTransform *iface, LONGLONG lower, LONGLONG upper)
-{
-    ok(0, "Unexpected %s call.\n", __func__);
-    return E_NOTIMPL;
-}
-
-static HRESULT WINAPI test_transform_ProcessEvent(IMFTransform *iface, DWORD id, IMFMediaEvent *event)
-{
-    ok(0, "Unexpected %s call.\n", __func__);
-    return E_NOTIMPL;
-}
-
-DEFINE_EXPECT(test_transform_ProcessMessage_BEGIN_STREAMING);
-DEFINE_EXPECT(test_transform_ProcessMessage_START_OF_STREAM);
-DEFINE_EXPECT(test_transform_ProcessMessage_FLUSH);
-
-static HRESULT WINAPI test_transform_ProcessMessage(IMFTransform *iface, MFT_MESSAGE_TYPE message, ULONG_PTR param)
-{
-    switch (message)
-    {
-    case MFT_MESSAGE_NOTIFY_BEGIN_STREAMING:
-        CHECK_EXPECT(test_transform_ProcessMessage_BEGIN_STREAMING);
-        add_object_state(&actual_object_state_record, MFT_BEGIN);
-        return S_OK;
-
-    case MFT_MESSAGE_NOTIFY_START_OF_STREAM:
-        CHECK_EXPECT(test_transform_ProcessMessage_START_OF_STREAM);
-        add_object_state(&actual_object_state_record, MFT_START);
-        return S_OK;
-
-    case MFT_MESSAGE_COMMAND_FLUSH:
-        CHECK_EXPECT(test_transform_ProcessMessage_FLUSH);
-        add_object_state(&actual_object_state_record, MFT_FLUSH);
-        return S_OK;
-
-    default:
-        ok(0, "Unexpected %s call %#x.\n", __func__, message);
-        return E_NOTIMPL;
-    }
-}
-
-DEFINE_EXPECT(test_transform_ProcessInput);
-DEFINE_EXPECT(test_transform_ProcessOutput);
-
-static HRESULT WINAPI test_transform_ProcessInput(IMFTransform *iface, DWORD id, IMFSample *sample, DWORD flags)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    HRESULT hr;
-
-    if (expect_test_transform_ProcessInput)
-    {
-        if (transform->output)
-        {
-            hr = MF_E_NOTACCEPTING;
-        }
-        else
-        {
-            IMFSample_AddRef(transform->output = sample);
-            hr = S_OK;
-        }
-    }
-    else
-    {
-        hr = E_NOTIMPL;
-    }
-
-    CHECK_EXPECT(test_transform_ProcessInput);
-    add_object_state(&actual_object_state_record, MFT_PROCESS_INPUT);
-
-    return hr;
-}
-
-static HRESULT WINAPI test_transform_ProcessOutput(IMFTransform *iface, DWORD flags, DWORD count,
-        MFT_OUTPUT_DATA_BUFFER *data, DWORD *status)
-{
-    struct test_transform *transform = test_transform_from_IMFTransform(iface);
-    HRESULT hr;
-
-    if (expect_test_transform_ProcessOutput)
-    {
-        if (transform->output)
-        {
-            *status = 0;
-            data->pSample = transform->output;
-            transform->output = NULL;
-            hr = S_OK;
-        }
-        else
-        {
-            hr = MF_E_TRANSFORM_NEED_MORE_INPUT;
-        }
-    }
-    else
-    {
-        hr = E_NOTIMPL;
-    }
-
-    CHECK_EXPECT2(test_transform_ProcessOutput);
-    add_object_state(&actual_object_state_record, MFT_PROCESS_OUTPUT);
-
-    return hr;
-}
-
-static const IMFTransformVtbl test_transform_vtbl =
-{
-    test_transform_QueryInterface,
-    test_transform_AddRef,
-    test_transform_Release,
-    test_transform_GetStreamLimits,
-    test_transform_GetStreamCount,
-    test_transform_GetStreamIDs,
-    test_transform_GetInputStreamInfo,
-    test_transform_GetOutputStreamInfo,
-    test_transform_GetAttributes,
-    test_transform_GetInputStreamAttributes,
-    test_transform_GetOutputStreamAttributes,
-    test_transform_DeleteInputStream,
-    test_transform_AddInputStreams,
-    test_transform_GetInputAvailableType,
-    test_transform_GetOutputAvailableType,
-    test_transform_SetInputType,
-    test_transform_SetOutputType,
-    test_transform_GetInputCurrentType,
-    test_transform_GetOutputCurrentType,
-    test_transform_GetInputStatus,
-    test_transform_GetOutputStatus,
-    test_transform_SetOutputBounds,
-    test_transform_ProcessEvent,
-    test_transform_ProcessMessage,
-    test_transform_ProcessInput,
-    test_transform_ProcessOutput,
-};
-
-static HRESULT WINAPI test_transform_create(UINT input_count, IMFMediaType **input_types,
-        UINT output_count, IMFMediaType **output_types, BOOL d3d_aware, IMFTransform **out)
-{
-    struct test_transform *transform;
-
-    if (!(transform = calloc(1, sizeof(*transform))))
-        return E_OUTOFMEMORY;
-    transform->IMFTransform_iface.lpVtbl = &test_transform_vtbl;
-    transform->refcount = 1;
-
-    transform->input_count = input_count;
-    transform->input_types = input_types;
-    transform->input_type = input_types[0];
-    IMFMediaType_AddRef(transform->input_type);
-    transform->output_count = output_count;
-    transform->output_types = output_types;
-    transform->output_type = output_types[0];
-    IMFMediaType_AddRef(transform->output_type);
-
-    *out = &transform->IMFTransform_iface;
-    return S_OK;
-}
-
-static void test_media_session_seek(void)
-{
-    static const struct object_state_record expected_start_state_records = {{SOURCE_START, MFT_START, SINK_ON_CLOCK_START}, 3};
-    static const struct object_state_record expected_sample_request_and_delivery_records = {{MFT_PROCESS_OUTPUT, SOURCE_REQUEST_SAMPLE, MFT_PROCESS_INPUT, MFT_PROCESS_OUTPUT, SINK_PROCESS_SAMPLE}, 5};
-    static const struct object_state_record expected_sample_request_only_records = {{MFT_PROCESS_OUTPUT, SOURCE_REQUEST_SAMPLE}, 2};
-    static const struct object_state_record expected_paused_state_records = {{SINK_ON_CLOCK_PAUSE, SOURCE_PAUSE}, 2};
-    static const struct object_state_record expected_seek_start_no_pending_request_records = {{SOURCE_STOP, MFT_FLUSH, SOURCE_START, SINK_FLUSH, SINK_ON_CLOCK_START}, 5};
-    static const struct object_state_record expected_seek_start_pending_request_records = {{SOURCE_STOP, MFT_FLUSH, SOURCE_START, MFT_PROCESS_OUTPUT, SOURCE_REQUEST_SAMPLE, SINK_FLUSH, SINK_ON_CLOCK_START}, 7};
-
-    IMFClockStateSink test_seek_clock_sink = {&test_seek_clock_sink_vtbl};
-    MFT_OUTPUT_STREAM_INFO output_stream_info = {0};
-    IMFPresentationClock *presentation_clock;
-    struct test_callback *test_callback;
-    struct test_media_sink *media_sink;
-    struct test_source *media_source;
-    struct test_handler *handler;
-    IMFAsyncCallback *callback;
-    IMFMediaSession *session;
-    IMFMediaSource *source;
-    IMFTopology *topology;
-    PROPVARIANT propvar;
-    IMFMediaType *type;
-    IMFTransform *mft;
-    IMFClock *clock;
-    UINT32 status;
-    HRESULT hr;
-    INT i;
-
-    handler = create_test_handler();
-    media_sink = create_test_media_sink(&handler->IMFMediaTypeHandler_iface);
-    IMFMediaTypeHandler_Release(&handler->IMFMediaTypeHandler_iface);
-
-    hr = MFStartup(MF_VERSION, MFSTARTUP_FULL);
-    ok(hr == S_OK, "Failed to start up, hr %#lx.\n", hr);
-
-    hr = MFCreateMediaSession(NULL, &session);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    source = create_test_source(TRUE);
-    media_source = impl_test_source_from_IMFMediaSource(source);
-    for (i = 0; i < media_source->stream_count; i++)
-        media_source->streams[i]->test_expect = TRUE;
-
-    hr = MFCreateMediaType(&type);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaType_SetGUID(type, &MF_MT_MAJOR_TYPE, &MFMediaType_Video);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaType_SetGUID(type, &MF_MT_SUBTYPE, &MFVideoFormat_RGB32);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaType_SetUINT64(type, &MF_MT_FRAME_SIZE, (UINT64)640 << 32 | 480);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    mft = NULL;
-    hr = test_transform_create(1, &type, 1, &type, FALSE, &mft);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    test_transform_set_output_stream_info(mft, &output_stream_info);
-    IMFMediaType_Release(type);
-
-    SET_EXPECT(test_transform_ProcessMessage_BEGIN_STREAMING);
-    topology = create_test_topology_unk(source, (IUnknown*)media_sink->stream, (IUnknown*) mft, NULL);
-    hr = IMFMediaSession_SetTopology(session, 0, topology);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    IMFTopology_Release(topology);
-
-    hr = IMFMediaSession_GetClock(session, &clock);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFClock_QueryInterface(clock, &IID_IMFPresentationClock, (void **)&presentation_clock);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFPresentationClock_AddClockStateSink(presentation_clock, &test_seek_clock_sink);
-    ok(hr == S_OK, "Failed to add a sink, hr %#lx.\n", hr);
-    IMFClock_Release(clock);
-
-    callback = create_test_callback(TRUE);
-    test_callback = impl_from_IMFAsyncCallback(callback);
-    PropVariantInit(&propvar);
-    hr = wait_media_event(session, callback, MESessionTopologyStatus, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaEvent_GetUINT32(test_callback->media_event, &MF_EVENT_TOPOLOGY_STATUS, &status);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(status == MF_TOPOSTATUS_READY, "Unexpected status %d.\n", status);
-    PropVariantClear(&propvar);
-    CHECK_CALLED(test_transform_ProcessMessage_BEGIN_STREAMING);
-
-    memset(&actual_object_state_record, 0, sizeof(actual_object_state_record));
-    SET_EXPECT(test_media_sink_GetPresentationClock);
-    SET_EXPECT(test_media_sink_SetPresentationClock);
-    SET_EXPECT(test_media_sink_GetStreamSinkCount);
-    SET_EXPECT(test_transform_ProcessMessage_START_OF_STREAM);
-    hr = IMFMediaSession_Start(session, NULL, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = wait_media_event(session, callback, MESessionTopologyStatus, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    hr = IMFMediaEvent_GetUINT32(test_callback->media_event, &MF_EVENT_TOPOLOGY_STATUS, &status);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(status == MF_TOPOSTATUS_STARTED_SOURCE, "Unexpected status %d.\n", status);
-    PropVariantClear(&propvar);
-
-    hr = IMFStreamSink_QueueEvent(media_sink->stream, MEStreamSinkStarted, &GUID_NULL, S_OK, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = wait_media_event(session, callback, MESessionStarted, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    Sleep(20);
-
-    todo_wine
-    CHECK_CALLED(test_media_sink_GetPresentationClock);
-    CHECK_CALLED(test_media_sink_SetPresentationClock);
-    todo_wine
-    CHECK_CALLED(test_media_sink_GetStreamSinkCount);
-    CHECK_CALLED(test_transform_ProcessMessage_START_OF_STREAM);
-
-    compare_object_states(&actual_object_state_record, &expected_start_state_records);
-
-    /* Test a sample request with sample delivery, then pause and then start with a seek */
-    memset(&actual_object_state_record, 0, sizeof(actual_object_state_record));
-    SET_EXPECT(test_media_stream_RequestSample);
-    SET_EXPECT(test_transform_ProcessOutput);
-    SET_EXPECT(test_transform_ProcessInput);
-    SET_EXPECT(test_stream_sink_ProcessSample);
-    IMFStreamSink_QueueEvent(media_sink->stream, MEStreamSinkRequestSample, &GUID_NULL, S_OK, &propvar);
-
-    Sleep(20);
-
-    CHECK_CALLED(test_media_stream_RequestSample);
-    CHECK_CALLED(test_transform_ProcessOutput);
-    CHECK_CALLED(test_transform_ProcessInput);
-    CHECK_CALLED(test_stream_sink_ProcessSample);
-
-    compare_object_states(&actual_object_state_record, &expected_sample_request_and_delivery_records);
-
-    SET_EXPECT(test_media_sink_GetPresentationClock);
-    SET_EXPECT(test_media_sink_GetStreamSinkCount);
-
-    PropVariantClear(&propvar);
-    hr = IMFMediaSession_Start(session, NULL, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = wait_media_event(session, callback, MESessionStarted, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    todo_wine
-    CHECK_CALLED(test_media_sink_GetPresentationClock);
-    todo_wine
-    CHECK_CALLED(test_media_sink_GetStreamSinkCount);
-
-    memset(&actual_object_state_record, 0, sizeof(actual_object_state_record));
-    hr = IMFMediaSession_Pause(session);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    Sleep(20);
-
-    hr = IMFStreamSink_QueueEvent(media_sink->stream, MEStreamSinkPaused, &GUID_NULL, S_OK, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = wait_media_event(session, callback, MESessionPaused, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    compare_object_states(&actual_object_state_record, &expected_paused_state_records);
-
-    memset(&actual_object_state_record, 0, sizeof(actual_object_state_record));
-    SET_EXPECT(test_media_sink_GetPresentationClock);
-    SET_EXPECT(test_media_sink_GetStreamSinkCount);
-    SET_EXPECT(test_stream_sink_Flush);
-    SET_EXPECT(test_transform_ProcessMessage_FLUSH);
-    propvar.vt = VT_I8;
-    propvar.hVal.QuadPart = 10000000;
-    hr = IMFMediaSession_Start(session, NULL, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    PropVariantClear(&propvar);
-
-    Sleep(20);
-
-    hr = IMFStreamSink_QueueEvent(media_sink->stream, MEStreamSinkStarted, &GUID_NULL, S_OK, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = wait_media_event(session, callback, MESessionStarted, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    todo_wine
-    CHECK_CALLED(test_media_sink_GetPresentationClock);
-    todo_wine
-    CHECK_CALLED(test_media_sink_GetStreamSinkCount);
-    CHECK_CALLED(test_stream_sink_Flush);
-    CHECK_CALLED(test_transform_ProcessMessage_FLUSH);
-
-    flaky
-    compare_object_states(&actual_object_state_record, &expected_seek_start_no_pending_request_records);
-
-    /* Test a sample request only (i.e. with no sample delivery), then pause and then start with a seek */
-    for (i = 0; i < media_source->stream_count; i++)
-        media_source->streams[i]->delay_sample = TRUE;
-
-    memset(&actual_object_state_record, 0, sizeof(actual_object_state_record));
-    SET_EXPECT(test_media_stream_RequestSample);
-    SET_EXPECT(test_transform_ProcessOutput);
-    IMFStreamSink_QueueEvent(media_sink->stream, MEStreamSinkRequestSample, &GUID_NULL, S_OK, &propvar);
-
-    Sleep(20);
-
-    CHECK_CALLED(test_media_stream_RequestSample);
-    CHECK_CALLED(test_transform_ProcessOutput);
-
-    compare_object_states(&actual_object_state_record, &expected_sample_request_only_records);
-
-    memset(&actual_object_state_record, 0, sizeof(actual_object_state_record));
-    hr = IMFMediaSession_Pause(session);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    Sleep(20);
-
-    hr = IMFStreamSink_QueueEvent(media_sink->stream, MEStreamSinkPaused, &GUID_NULL, S_OK, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = wait_media_event(session, callback, MESessionPaused, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    compare_object_states(&actual_object_state_record, &expected_paused_state_records);
-
-    memset(&actual_object_state_record, 0, sizeof(actual_object_state_record));
-    SET_EXPECT(test_media_sink_GetPresentationClock);
-    SET_EXPECT(test_media_sink_GetStreamSinkCount);
-    SET_EXPECT(test_stream_sink_Flush);
-    SET_EXPECT(test_transform_ProcessMessage_FLUSH);
-    SET_EXPECT(test_transform_ProcessOutput);
-    SET_EXPECT(test_media_stream_RequestSample);
-    propvar.vt = VT_I8;
-    propvar.hVal.QuadPart = 10000000;
-    hr = IMFMediaSession_Start(session, NULL, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    PropVariantClear(&propvar);
-
-    Sleep(20);
-
-    hr = IMFStreamSink_QueueEvent(media_sink->stream, MEStreamSinkStarted, &GUID_NULL, S_OK, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    hr = wait_media_event(session, callback, MESessionStarted, 1000, &propvar);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    todo_wine
-    CHECK_CALLED(test_media_sink_GetPresentationClock);
-    todo_wine
-    CHECK_CALLED(test_media_sink_GetStreamSinkCount);
-    CHECK_CALLED(test_stream_sink_Flush);
-    CHECK_CALLED(test_transform_ProcessMessage_FLUSH);
-    CHECK_CALLED(test_transform_ProcessOutput);
-    CHECK_CALLED(test_media_stream_RequestSample);
-
-    flaky
-    compare_object_states(&actual_object_state_record, &expected_seek_start_pending_request_records);
-
-    IMFPresentationClock_RemoveClockStateSink(presentation_clock, &test_seek_clock_sink);
-    IMFPresentationClock_Release(presentation_clock);
-    IMFAsyncCallback_Release(callback);
-    IMFTransform_Release(mft);
-
-    hr = IMFMediaSession_Shutdown(session);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-    ok(media_sink->shutdown, "Media sink didn't shutdown.\n");
-
-    hr = IMFMediaSource_Shutdown(source);
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
-
-    IMFMediaSession_Release(session);
-    IMFMediaSource_Release(source);
-    IMFMediaSink_Release(&media_sink->IMFMediaSink_iface);
-
-    hr = MFShutdown();
-    ok(hr == S_OK, "Unexpected hr %#lx.\n", hr);
 }
 
 START_TEST(mf)
@@ -8468,11 +7161,9 @@
     test_MFRequireProtectedEnvironment();
     test_mpeg4_media_sink();
     test_MFCreateSequencerSegmentOffset();
-    test_h264_output_alignment();
     test_media_session_Start();
     test_MFEnumDeviceSources();
     test_media_session_Close();
     test_media_session_source_shutdown();
     test_media_session_thinning();
-    test_media_session_seek();
-}
+}
