--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -29,16 +29,15 @@
 #define WIN32_NO_STATUS
 #include "ntgdi_private.h"
 #include "ntuser_private.h"
-#include "wine/opengl_driver.h"
 #include "wine/server.h"
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(win);
 
+#define NB_USER_HANDLES  ((LAST_USER_HANDLE - FIRST_USER_HANDLE + 1) >> 1)
 #define USER_HANDLE_TO_INDEX(hwnd) ((LOWORD(hwnd) - FIRST_USER_HANDLE) >> 1)
-#define USER_HANDLE_FROM_INDEX(index, generation) UlongToHandle( (index << 1) + FIRST_USER_HANDLE + (generation << 16) )
 
-static void *client_objects[MAX_USER_HANDLES];
+static void *user_handles[NB_USER_HANDLES];
 
 #define SWP_AGG_NOGEOMETRYCHANGE \
     (SWP_NOSIZE | SWP_NOCLIENTSIZE | SWP_NOZORDER)
@@ -53,39 +52,15 @@
 #define PLACE_MAX		0x0002
 #define PLACE_RECT		0x0004
 
-static volatile unsigned int startup_info_flags;
-static unsigned int startup_show_window;
-
-static unsigned int set_startup_info_flags( unsigned int mask, unsigned int flags )
-{
-    unsigned int prev, new;
-
-    do
-    {
-        prev = startup_info_flags;
-        new = (prev & ~mask) | flags;
-    } while (InterlockedCompareExchange( (LONG volatile *)&startup_info_flags, new, prev ) != prev );
-    return prev;
-}
-
-void init_startup_info(void)
-{
-    RTL_USER_PROCESS_PARAMETERS *p = NtCurrentTeb()->Peb->ProcessParameters;
-
-    startup_show_window = p->wShowWindow;
-    set_startup_info_flags( ~0u, p->dwFlags );
-}
-
 /***********************************************************************
  *           alloc_user_handle
  */
-HANDLE alloc_user_handle( void *ptr, unsigned short type )
+HANDLE alloc_user_handle( struct user_object *ptr, unsigned int type )
 {
     HANDLE handle = 0;
 
     SERVER_START_REQ( alloc_user_handle )
     {
-        req->type = type;
         if (!wine_server_call_err( req )) handle = wine_server_ptr_handle( reply->handle );
     }
     SERVER_END_REQ;
@@ -94,149 +69,66 @@
     {
         UINT index = USER_HANDLE_TO_INDEX( handle );
 
-        assert( index < MAX_USER_HANDLES );
-        InterlockedExchangePointer( &client_objects[index], ptr );
+        assert( index < NB_USER_HANDLES );
+        ptr->handle = handle;
+        ptr->type = type;
+        InterlockedExchangePointer( &user_handles[index], ptr );
     }
     return handle;
 }
 
-static BOOL is_valid_entry_uniq( HANDLE handle, unsigned short type, UINT64 uniq )
-{
-    if (!HIWORD(handle) || HIWORD(handle) == 0xffff) return LOWORD(uniq) == type;
-    return uniq == MAKELONG(type, HIWORD(handle));
-}
-
-static BOOL is_valid_entry( HANDLE handle, unsigned short type )
-{
-    const volatile struct user_entry *entries = shared_session->user_entries;
-    WORD index = USER_HANDLE_TO_INDEX( handle );
-    if (index >= MAX_USER_HANDLES) return FALSE;
-    return is_valid_entry_uniq( handle, type, ReadAcquire64( (LONG64 *)&entries[index].uniq ) );
-}
-
-static BOOL read_acquire_user_entry( HANDLE handle, unsigned short type, const volatile struct user_entry *src, struct user_entry *dst )
-{
-    UINT64 uniq = ReadAcquire64( (LONG64 *)&src->uniq );
-    if (!is_valid_entry_uniq( handle, type, uniq )) return FALSE;
-    dst->offset = src->offset;
-    dst->tid = src->tid;
-    dst->pid = src->pid;
-    dst->id = src->id;
-    __SHARED_READ_FENCE;
-    dst->uniq = ReadNoFence64( &src->uniq );
-    return dst->uniq == uniq;
-}
-
-static BOOL get_user_entry( HANDLE handle, unsigned short type, struct user_entry *entry, HANDLE *full )
-{
-    const volatile struct user_entry *entries = shared_session->user_entries;
-    WORD index = USER_HANDLE_TO_INDEX( handle );
-
-    if (index >= MAX_USER_HANDLES) return FALSE;
-    if (!read_acquire_user_entry( handle, type, entries + index, entry )) return FALSE;
-    *full = USER_HANDLE_FROM_INDEX( index, entry->generation );
-    return TRUE;
-}
-
-static BOOL get_user_entry_at( WORD index, unsigned short type, struct user_entry *entry, HANDLE *full )
-{
-    const volatile struct user_entry *entries = shared_session->user_entries;
-    if (index >= MAX_USER_HANDLES) return FALSE;
-    if (!read_acquire_user_entry( 0, type, entries + index, entry )) return FALSE;
-    *full = USER_HANDLE_FROM_INDEX( index, entry->generation );
-    return TRUE;
-}
-
-static NTSTATUS get_shared_window( HANDLE handle, struct object_lock *lock, const window_shm_t **window_shm )
-{
-    const shared_object_t *object;
-    struct user_entry entry;
-
-    if (!get_user_entry( handle, NTUSER_OBJ_WINDOW, &entry, &handle )) return STATUS_INVALID_HANDLE;
-
-    if (lock->id) object = CONTAINING_RECORD( *window_shm, shared_object_t, shm.window );
-    else
-    {
-        object = find_shared_session_object( entry.id, entry.offset );
-        if (!object) return STATUS_INVALID_HANDLE;
-    }
-
-    if (!lock->id || !shared_object_release_seqlock( object, lock->seq ))
-    {
-        shared_object_acquire_seqlock( object, &lock->seq );
-        *window_shm = &object->shm.window;
-        lock->id = object->id;
-        return STATUS_PENDING;
-    }
-
-    return STATUS_SUCCESS;
-}
-
-struct obj_locator get_window_class_locator( HWND hwnd )
-{
-    struct object_lock lock = OBJECT_LOCK_INIT;
-    const window_shm_t *window_shm = NULL;
-    struct obj_locator locator = {0};
-    NTSTATUS status;
-
-    while ((status = get_shared_window( hwnd, &lock, &window_shm )) == STATUS_PENDING)
-        locator = window_shm->class;
-    if (status) memset( &locator, 0, sizeof(locator) );
-
-    return locator;
-}
-
 /***********************************************************************
  *           get_user_handle_ptr
  */
-void *get_user_handle_ptr( HANDLE handle, unsigned short type )
+void *get_user_handle_ptr( HANDLE handle, unsigned int type )
 {
+    struct user_object *ptr;
     WORD index = USER_HANDLE_TO_INDEX( handle );
-    void *ptr = NULL;
-    struct user_entry entry;
 
-    if (index >= MAX_USER_HANDLES) return NULL;
+    if (index >= NB_USER_HANDLES) return NULL;
 
     user_lock();
-
-    if (!get_user_entry( handle, type, &entry, &handle )) ptr = NULL;
-    else if (entry.pid != GetCurrentProcessId()) ptr = OBJ_OTHER_PROCESS;
-    else ptr = client_objects[index];
-
-    if (!ptr || ptr == OBJ_OTHER_PROCESS) user_unlock();
+    if ((ptr = user_handles[index]))
+    {
+        if (ptr->type == type &&
+            ((UINT)(UINT_PTR)ptr->handle == (UINT)(UINT_PTR)handle ||
+             !HIWORD(handle) || HIWORD(handle) == 0xffff))
+            return ptr;
+        ptr = NULL;
+    }
+    else ptr = OBJ_OTHER_PROCESS;
+    user_unlock();
     return ptr;
 }
 
 /***********************************************************************
- *           next_thread_user_object
+ *           next_process_user_handle_ptr
  *
  * user_lock must be held by caller.
  */
-void *next_thread_user_object( UINT tid, HANDLE *handle, unsigned short type )
+void *next_process_user_handle_ptr( HANDLE *handle, unsigned int type )
 {
+    struct user_object *ptr;
     WORD index = *handle ? USER_HANDLE_TO_INDEX( *handle ) + 1 : 0;
-    struct user_entry entry;
-    UINT i;
 
-    for (i = index; i < MAX_USER_HANDLES; i++)
+    while (index < NB_USER_HANDLES)
     {
-        if (!get_user_entry_at( i, type, &entry, handle )) continue;
-        if (entry.pid != GetCurrentProcessId()) continue;
-        if (tid != -1 && entry.tid != tid) continue;
-        return client_objects[i];
+        if (!(ptr = user_handles[index++])) continue;  /* OBJ_OTHER_PROCESS */
+        if (ptr->type != type) continue;
+        *handle = ptr->handle;
+        return ptr;
     }
-
     return NULL;
 }
 
 /***********************************************************************
  *           set_user_handle_ptr
  */
-static void set_user_handle_ptr( HANDLE handle, void *ptr )
+static void set_user_handle_ptr( HANDLE handle, struct user_object *ptr )
 {
     WORD index = USER_HANDLE_TO_INDEX(handle);
-    assert( index < MAX_USER_HANDLES );
-    InterlockedExchangePointer( &client_objects[index], ptr );
+    assert( index < NB_USER_HANDLES );
+    InterlockedExchangePointer( &user_handles[index], ptr );
 }
 
 /***********************************************************************
@@ -251,19 +143,18 @@
 /***********************************************************************
  *           free_user_handle
  */
-void *free_user_handle( HANDLE handle, unsigned short type )
+void *free_user_handle( HANDLE handle, unsigned int type )
 {
-    void *ptr;
+    struct user_object *ptr;
     WORD index = USER_HANDLE_TO_INDEX( handle );
 
     if ((ptr = get_user_handle_ptr( handle, type )) && ptr != OBJ_OTHER_PROCESS)
     {
         SERVER_START_REQ( free_user_handle )
         {
-            req->type = type;
             req->handle = wine_server_user_handle( handle );
             if (wine_server_call( req )) ptr = NULL;
-            else InterlockedCompareExchangePointer( &client_objects[index], NULL, ptr );
+            else InterlockedCompareExchangePointer( &user_handles[index], NULL, ptr );
         }
         SERVER_END_REQ;
         user_unlock();
@@ -271,131 +162,6 @@
     return ptr;
 }
 
-static pthread_mutex_t surfaces_lock = PTHREAD_MUTEX_INITIALIZER;
-static struct list client_surfaces = LIST_INIT( client_surfaces );
-
-void detach_client_surfaces( HWND hwnd )
-{
-    struct list detached = LIST_INIT( detached );
-    struct client_surface *surface, *next;
-
-    pthread_mutex_lock( &surfaces_lock );
-
-    LIST_FOR_EACH_ENTRY_SAFE( surface, next, &client_surfaces, struct client_surface, entry )
-    {
-        if (surface->hwnd != hwnd) continue;
-
-        list_remove( &surface->entry );
-        list_add_tail( &detached, &surface->entry );
-        client_surface_add_ref( surface );
-
-        surface->funcs->detach( surface );
-        surface->hwnd = NULL;
-    }
-
-    pthread_mutex_unlock( &surfaces_lock );
-
-    LIST_FOR_EACH_ENTRY_SAFE( surface, next, &detached, struct client_surface, entry )
-    {
-        list_remove( &surface->entry );
-        client_surface_release( surface );
-    }
-}
-
-static void update_client_surfaces( HWND hwnd )
-{
-    struct client_surface *surface, *next;
-
-    pthread_mutex_lock( &surfaces_lock );
-
-    LIST_FOR_EACH_ENTRY_SAFE( surface, next, &client_surfaces, struct client_surface, entry )
-    {
-        if (NtUserGetAncestor( surface->hwnd, GA_ROOT ) != hwnd) continue;
-        surface->funcs->update( surface );
-        InterlockedExchange( &surface->updated, 1 );
-    }
-
-    pthread_mutex_unlock( &surfaces_lock );
-}
-
-void *client_surface_create( UINT size, const struct client_surface_funcs *funcs, HWND hwnd )
-{
-    struct client_surface *surface;
-
-    if (!(surface = calloc( 1, size ))) return NULL;
-    surface->funcs = funcs;
-    surface->ref = 1;
-    surface->hwnd = hwnd;
-
-    pthread_mutex_lock( &surfaces_lock );
-    list_add_tail( &client_surfaces, &surface->entry );
-    pthread_mutex_unlock( &surfaces_lock );
-
-    TRACE( "created %s\n", debugstr_client_surface( surface ) );
-    return surface;
-}
-
-void client_surface_add_ref( struct client_surface *surface )
-{
-    ULONG ref = InterlockedIncrement( &surface->ref );
-    TRACE( "%s increasing refcount to %u\n", debugstr_client_surface( surface ), ref );
-}
-
-void client_surface_release( struct client_surface *surface )
-{
-    ULONG ref = InterlockedDecrement( &surface->ref );
-    TRACE( "%s decreasing refcount to %u\n", debugstr_client_surface( surface ), ref );
-
-    if (!ref)
-    {
-        pthread_mutex_lock( &surfaces_lock );
-        if (surface->hwnd)
-        {
-            surface->funcs->detach( surface );
-            list_remove( &surface->entry );
-        }
-        pthread_mutex_unlock( &surfaces_lock );
-
-        surface->funcs->destroy( surface );
-        free( surface );
-    }
-}
-
-void client_surface_present( struct client_surface *surface )
-{
-    HDC hdc = 0;
-    HWND hwnd;
-
-    pthread_mutex_lock( &surfaces_lock );
-    if ((hwnd = surface->hwnd))
-    {
-        if (surface->offscreen) hdc = NtUserGetDCEx( hwnd, 0, DCX_CACHE | DCX_USESTYLE );
-        surface->funcs->present( surface, hdc );
-        if (hdc) NtUserReleaseDC( hwnd, hdc );
-    }
-    pthread_mutex_unlock( &surfaces_lock );
-}
-
-void client_surface_update( struct client_surface *surface )
-{
-    pthread_mutex_lock( &surfaces_lock );
-    if (surface->hwnd) surface->funcs->update( surface );
-    pthread_mutex_unlock( &surfaces_lock );
-}
-
-BOOL is_client_surface_window( struct client_surface *surface, HWND hwnd )
-{
-    BOOL ret;
-
-    if (!surface) return FALSE;
-    pthread_mutex_lock( &surfaces_lock );
-    if (hwnd) ret = surface->hwnd == hwnd;
-    else ret = surface->hwnd != NULL;
-    pthread_mutex_unlock( &surfaces_lock );
-
-    return ret;
-}
-
 /*******************************************************************
  *           get_hwnd_message_parent
  *
@@ -416,16 +182,36 @@
  */
 HWND get_full_window_handle( HWND hwnd )
 {
-    struct user_entry entry;
-    HANDLE handle;
+    WND *win;
 
     if (!hwnd || (ULONG_PTR)hwnd >> 16) return hwnd;
     if (LOWORD(hwnd) <= 1 || LOWORD(hwnd) == 0xffff) return hwnd;
     /* do sign extension for -2 and -3 */
     if (LOWORD(hwnd) >= (WORD)-3) return (HWND)(LONG_PTR)(INT16)LOWORD(hwnd);
 
-    if (!get_user_entry( hwnd, NTUSER_OBJ_WINDOW, &entry, &handle )) return hwnd;
-    return handle;
+    if (!(win = get_win_ptr( hwnd ))) return hwnd;
+
+    if (win == WND_DESKTOP)
+    {
+        if (LOWORD(hwnd) == LOWORD(get_desktop_window())) return get_desktop_window();
+        else return get_hwnd_message_parent();
+    }
+
+    if (win != WND_OTHER_PROCESS)
+    {
+        hwnd = win->obj.handle;
+        release_win_ptr( win );
+    }
+    else  /* may belong to another process */
+    {
+        SERVER_START_REQ( get_window_info )
+        {
+            req->handle = wine_server_user_handle( hwnd );
+            if (!wine_server_call_err( req )) hwnd = wine_server_ptr_handle( reply->full_handle );
+        }
+        SERVER_END_REQ;
+    }
+    return hwnd;
 }
 
 /*******************************************************************
@@ -474,12 +260,14 @@
  */
 HWND is_current_thread_window( HWND hwnd )
 {
-    struct user_entry entry;
-    HANDLE handle;
+    WND *win;
+    HWND ret = 0;
 
-    if (!get_user_entry( hwnd, NTUSER_OBJ_WINDOW, &entry, &handle )) return 0;
-    if (entry.tid != GetCurrentThreadId()) return 0;
-    return handle;
+    if (!(win = get_win_ptr( hwnd )) || win == WND_OTHER_PROCESS || win == WND_DESKTOP)
+        return 0;
+    if (win->tid == GetCurrentThreadId()) ret = win->obj.handle;
+    release_win_ptr( win );
+    return ret;
 }
 
 /***********************************************************************
@@ -489,40 +277,73 @@
  */
 HWND is_current_process_window( HWND hwnd )
 {
-    struct user_entry entry;
-    HANDLE handle;
+    WND *ptr;
+    HWND ret;
 
-    if (!get_user_entry( hwnd, NTUSER_OBJ_WINDOW, &entry, &handle )) return 0;
-    if (entry.pid != GetCurrentProcessId()) return 0;
-    return handle;
+    if (!(ptr = get_win_ptr( hwnd )) || ptr == WND_OTHER_PROCESS || ptr == WND_DESKTOP) return 0;
+    ret = ptr->obj.handle;
+    release_win_ptr( ptr );
+    return ret;
 }
 
 /* see IsWindow */
 BOOL is_window( HWND hwnd )
 {
-    if (!hwnd) return FALSE;
-    if (!is_valid_entry( hwnd, NTUSER_OBJ_WINDOW ))
+    WND *win;
+    BOOL ret;
+
+    if (!(win = get_win_ptr( hwnd ))) return FALSE;
+    if (win == WND_DESKTOP) return TRUE;
+
+    if (win != WND_OTHER_PROCESS)
     {
-        RtlSetLastWin32Error( ERROR_INVALID_WINDOW_HANDLE );
-        return FALSE;
+        release_win_ptr( win );
+        return TRUE;
     }
-    return TRUE;
+
+    /* check other processes */
+    SERVER_START_REQ( get_window_info )
+    {
+        req->handle = wine_server_user_handle( hwnd );
+        ret = !wine_server_call_err( req );
+    }
+    SERVER_END_REQ;
+    return ret;
 }
 
 /* see GetWindowThreadProcessId */
 DWORD get_window_thread( HWND hwnd, DWORD *process )
 {
-    struct user_entry entry;
-    HANDLE handle;
+    WND *ptr;
+    DWORD tid = 0;
 
-    if (!get_user_entry( hwnd, NTUSER_OBJ_WINDOW, &entry, &handle ))
+    if (!(ptr = get_win_ptr( hwnd )))
     {
-        RtlSetLastWin32Error( ERROR_INVALID_WINDOW_HANDLE );
+        RtlSetLastWin32Error( ERROR_INVALID_WINDOW_HANDLE);
         return 0;
     }
 
-    if (process) *process = entry.pid;
-    return entry.tid;
+    if (ptr != WND_OTHER_PROCESS && ptr != WND_DESKTOP)
+    {
+        /* got a valid window */
+        tid = ptr->tid;
+        if (process) *process = GetCurrentProcessId();
+        release_win_ptr( ptr );
+        return tid;
+    }
+
+    /* check other processes */
+    SERVER_START_REQ( get_window_info )
+    {
+        req->handle = wine_server_user_handle( hwnd );
+        if (!wine_server_call_err( req ))
+        {
+            tid = (DWORD)reply->tid;
+            if (process) *process = (DWORD)reply->pid;
+        }
+    }
+    SERVER_END_REQ;
+    return tid;
 }
 
 /* see GetParent */
@@ -539,7 +360,7 @@
     if (win == WND_DESKTOP) return 0;
     if (win == WND_OTHER_PROCESS)
     {
-        DWORD style = get_window_long( hwnd, GWL_STYLE );
+        LONG style = get_window_long( hwnd, GWL_STYLE );
         if (style & (WS_POPUP | WS_CHILD))
         {
             SERVER_START_REQ( get_window_tree )
@@ -618,8 +439,6 @@
     win = get_win_ptr( hwnd );
     if (!win || win == WND_OTHER_PROCESS || win == WND_DESKTOP) return 0;
 
-    if (parent && parent != NtUserGetDesktopWindow()) win->has_icons = FALSE;
-
     get_window_rect_rel( hwnd, COORDS_PARENT, &window_rect, get_dpi_for_window(hwnd) );
     get_window_rect_rel( hwnd, COORDS_SCREEN, &old_screen_rect, 0 );
 
@@ -631,6 +450,7 @@
         {
             old_parent = wine_server_ptr_handle( reply->old_parent );
             win->parent = parent = wine_server_ptr_handle( reply->full_parent );
+            win->dpi_context = reply->dpi_context;
         }
 
     }
@@ -647,8 +467,8 @@
     old_toplevel = NtUserGetAncestor( old_parent, GA_ROOT );
     if (new_toplevel != old_toplevel)
     {
-        if (new_toplevel) update_window_state( new_toplevel );
-        if (old_toplevel) update_window_state( old_toplevel );
+        update_window_state( new_toplevel );
+        update_window_state( old_toplevel );
     }
 
     winpos.hwnd = hwnd;
@@ -662,6 +482,17 @@
     set_window_pos( &winpos, new_screen_rect.left - old_screen_rect.left,
                     new_screen_rect.top - old_screen_rect.top );
 
+    {
+        WCHAR name[32];
+        UNICODE_STRING us = { 0, sizeof(name), name };
+
+        if (NtUserGetClassName( hwnd, FALSE, &us ) && ( !wcscmp( us.Buffer, u"SyberiaRenderWindowClass" ) || !wcscmp( us.Buffer, u"Post MortemRenderWindowClass" ) ))
+        {
+            ERR( "HACK: Hiding window.\n" );
+            was_visible = FALSE;
+        }
+    }
+
     if (was_visible) NtUserShowWindow( hwnd, SW_SHOW );
 
     set_thread_dpi_awareness_context( context );
@@ -673,27 +504,6 @@
 {
     HWND retval = 0;
 
-    if (rel == GW_ENABLEDPOPUP)
-    {
-        HWND *list;
-        int i;
-
-        if (NtUserGetAncestor( hwnd, GA_ROOT ) != hwnd) return NULL;
-        if (!(list = list_window_children( 0 ))) return NULL;
-
-        for (i = 0; !retval && list[i]; i++)
-        {
-            if (get_window_relative( list[i], GW_OWNER ) != hwnd) continue;
-            if (!is_window_visible( list[i] )) continue;
-            if (!is_window_enabled( list[i] )) continue;
-            retval = list[i];
-        }
-
-        free(list);
-
-        return retval;
-    }
-
     if (rel == GW_OWNER)  /* this one may be available locally */
     {
         WND *win = get_win_ptr( hwnd );
@@ -953,7 +763,7 @@
     HWND *list;
     BOOL retval = TRUE;
     int i;
-    DWORD style = get_window_long( hwnd, GWL_STYLE );
+    LONG style = get_window_long( hwnd, GWL_STYLE );
 
     if (!(style & WS_VISIBLE)) return FALSE;
     if ((style & WS_MINIMIZE) && icon && get_class_long_ptr( hwnd, GCLP_HICON, FALSE ))  return FALSE;
@@ -997,10 +807,8 @@
     return ret;
 }
 
-/*****************************************************************
- *           NtUserEnableWindow (win32u.@)
- */
-BOOL WINAPI NtUserEnableWindow( HWND hwnd, BOOL enable )
+/* see EnableWindow */
+BOOL enable_window( HWND hwnd, BOOL enable )
 {
     BOOL ret;
 
@@ -1014,7 +822,7 @@
 
     if (enable)
     {
-        ret = (set_window_style_bits( hwnd, 0, WS_DISABLED ) & WS_DISABLED) != 0;
+        ret = (set_window_style( hwnd, 0, WS_DISABLED ) & WS_DISABLED) != 0;
         if (ret)
         {
             NtUserNotifyWinEvent( EVENT_OBJECT_STATECHANGE, hwnd, OBJID_WINDOW, 0 );
@@ -1026,7 +834,7 @@
     {
         send_message( hwnd, WM_CANCELMODE, 0, 0 );
 
-        ret = (set_window_style_bits( hwnd, WS_DISABLED, 0 ) & WS_DISABLED) != 0;
+        ret = (set_window_style( hwnd, WS_DISABLED, 0 ) & WS_DISABLED) != 0;
         if (!ret)
         {
             NtUserNotifyWinEvent( EVENT_OBJECT_STATECHANGE, hwnd, OBJID_WINDOW, 0 );
@@ -1058,7 +866,7 @@
 /* see IsWindowEnabled */
 BOOL is_window_enabled( HWND hwnd )
 {
-    DWORD ret;
+    LONG ret;
 
     RtlSetLastWin32Error( NO_ERROR );
     ret = get_window_long( hwnd, GWL_STYLE );
@@ -1069,26 +877,64 @@
 /* see GetWindowDpiAwarenessContext */
 UINT get_window_dpi_awareness_context( HWND hwnd )
 {
-    struct object_lock lock = OBJECT_LOCK_INIT;
-    const window_shm_t *window_shm = NULL;
-    UINT status, ctx = 0;
-
-    while ((status = get_shared_window( hwnd, &lock, &window_shm )) == STATUS_PENDING)
-        ctx = window_shm->dpi_context;
-    if (status)
+    UINT ret = 0;
+    WND *win;
+
+    if (!(win = get_win_ptr( hwnd )))
     {
         RtlSetLastWin32Error( ERROR_INVALID_WINDOW_HANDLE );
         return 0;
     }
-
-    return ctx;
+    if (win == WND_DESKTOP) return NTUSER_DPI_PER_MONITOR_AWARE;
+    if (win != WND_OTHER_PROCESS)
+    {
+        ret = win->dpi_context;
+        release_win_ptr( win );
+    }
+    else
+    {
+        SERVER_START_REQ( get_window_info )
+        {
+            req->handle = wine_server_user_handle( hwnd );
+            if (!wine_server_call_err( req )) ret = reply->dpi_context;
+        }
+        SERVER_END_REQ;
+    }
+    return ret;
 }
 
 /* see GetDpiForWindow */
 UINT get_dpi_for_window( HWND hwnd )
 {
-    UINT raw_dpi, context = get_window_dpi_awareness_context( hwnd );
-    if (NTUSER_DPI_CONTEXT_IS_MONITOR_AWARE( context )) return get_win_monitor_dpi( hwnd, &raw_dpi );
+    WND *win;
+    UINT context = 0;
+
+    if (!(win = get_win_ptr( hwnd )))
+    {
+        RtlSetLastWin32Error( ERROR_INVALID_WINDOW_HANDLE );
+        return 0;
+    }
+    if (win == WND_DESKTOP)
+    {
+        RECT rect = {0};
+        return monitor_dpi_from_rect( rect, get_thread_dpi(), NULL, NULL );
+    }
+    if (win != WND_OTHER_PROCESS)
+    {
+        context = win->dpi_context;
+        release_win_ptr( win );
+    }
+    else
+    {
+        SERVER_START_REQ( get_window_info )
+        {
+            req->handle = wine_server_user_handle( hwnd );
+            if (!wine_server_call_err( req )) context = reply->dpi_context;
+        }
+        SERVER_END_REQ;
+    }
+
+    if (NTUSER_DPI_CONTEXT_IS_MONITOR_AWARE( context )) return get_win_monitor_dpi( hwnd, NULL, NULL );
     return NTUSER_DPI_CONTEXT_GET_DPI( context );
 }
 
@@ -1205,12 +1051,27 @@
             RtlSetLastWin32Error( ERROR_ACCESS_DENIED );
             return 0;
         }
-        SERVER_START_REQ( get_window_info )
+        SERVER_START_REQ( set_window_info )
         {
             req->handle = wine_server_user_handle( hwnd );
-            req->offset = offset;
-            req->size = size;
-            if (!wine_server_call_err( req )) retval = reply->info;
+            req->flags  = 0;  /* don't set anything, just retrieve */
+            req->extra_offset = (offset >= 0) ? offset : -1;
+            req->extra_size = (offset >= 0) ? size : 0;
+            if (!wine_server_call_err( req ))
+            {
+                switch(offset)
+                {
+                case GWL_STYLE:      retval = reply->old_style; break;
+                case GWL_EXSTYLE:    retval = reply->old_ex_style; break;
+                case GWLP_ID:        retval = reply->old_id; break;
+                case GWLP_HINSTANCE: retval = (ULONG_PTR)wine_server_get_ptr( reply->old_instance ); break;
+                case GWLP_USERDATA:  retval = reply->old_user_data; break;
+                default:
+                    if (offset >= 0) retval = get_win_data( &reply->old_extra_value, size );
+                    else RtlSetLastWin32Error( ERROR_INVALID_INDEX );
+                    break;
+                }
+            }
         }
         SERVER_END_REQ;
         return retval;
@@ -1281,7 +1142,12 @@
     return get_window_long_size( hwnd, offset, sizeof(WORD), TRUE );
 }
 
-UINT set_window_style_bits( HWND hwnd, UINT set_bits, UINT clear_bits )
+/***********************************************************************
+ *           set_window_style
+ *
+ * Change the style of a window.
+ */
+ULONG set_window_style( HWND hwnd, ULONG set_bits, ULONG clear_bits )
 {
     BOOL ok, made_visible = FALSE;
     STYLESTRUCT style;
@@ -1304,11 +1170,12 @@
     SERVER_START_REQ( set_window_info )
     {
         req->handle = wine_server_user_handle( hwnd );
-        req->offset = GWL_STYLE;
-        req->new_info = style.styleNew;
+        req->flags  = SET_WIN_STYLE;
+        req->style  = style.styleNew;
+        req->extra_offset = -1;
         if ((ok = !wine_server_call( req )))
         {
-            style.styleOld = reply->old_info;
+            style.styleOld = reply->old_style;
             win->dwStyle = style.styleNew;
         }
     }
@@ -1329,17 +1196,6 @@
     return style.styleOld;
 }
 
-/**********************************************************************
- *           NtUserAlterWindowStyle (win32u.@)
- */
-ULONG WINAPI NtUserAlterWindowStyle( HWND hwnd, UINT mask, UINT style )
-{
-    TRACE( "hwnd %p, mask %#x, style %#x\n", hwnd, mask, style );
-    /* FIXME: WS_TABSTOP shouldn't be there but we need it for BM_SETCHECK */
-    mask &= WS_TABSTOP | WS_VSCROLL | WS_HSCROLL | 0x23f;
-    return !!set_window_style_bits( hwnd, style & mask, mask & ~style );
-}
-
 static DWORD fix_exstyle( DWORD style, DWORD exstyle )
 {
     if ((exstyle & WS_EX_DLGMODALFRAME) ||
@@ -1380,6 +1236,8 @@
 /* Helper function for SetWindowLong(). */
 LONG_PTR set_window_long( HWND hwnd, INT offset, UINT size, LONG_PTR newval, BOOL ansi )
 {
+    const char *sgi = getenv( "SteamGameId" );
+
     BOOL ok, made_visible = FALSE, layered = FALSE;
     LONG_PTR retval = 0;
     STYLESTRUCT style;
@@ -1435,6 +1293,10 @@
         if (win->dwStyle & WS_MINIMIZE) newval |= WS_MINIMIZE;
         break;
     case GWL_EXSTYLE:
+        /* FIXME: Layered windows don't work well right now, disable them */
+        if (sgi && !strcmp( sgi, "694280" )) newval &= ~WS_EX_LAYERED;
+        if (sgi && !strcmp( sgi, "312670" )) newval &= ~WS_EX_LAYERED;
+        if (sgi && !strcmp( sgi, "700600" )) newval &= ~WS_EX_LAYERED;
         style.styleOld = win->dwExStyle;
         style.styleNew = newval;
         release_win_ptr( win );
@@ -1493,45 +1355,74 @@
         break;
     }
 
-    if (offset == GWLP_WNDPROC) newval = !!(win->flags & WIN_ISUNICODE);
-    if (offset == GWLP_USERDATA && size == sizeof(WORD)) newval = MAKELONG( newval, win->userdata >> 16 );
-
     SERVER_START_REQ( set_window_info )
     {
         req->handle = wine_server_user_handle( hwnd );
-        req->offset = offset;
-        req->new_info = newval;
-        req->size = size;
+        req->extra_offset = -1;
+        switch(offset)
+        {
+        case GWL_STYLE:
+            req->flags = SET_WIN_STYLE | SET_WIN_EXSTYLE;
+            req->style = newval;
+            req->ex_style = fix_exstyle(newval, win->dwExStyle);
+            break;
+        case GWL_EXSTYLE:
+            req->flags = SET_WIN_EXSTYLE;
+            req->ex_style = newval;
+            break;
+        case GWLP_ID:
+            req->flags = SET_WIN_ID;
+            req->extra_value = newval;
+            break;
+        case GWLP_HINSTANCE:
+            req->flags = SET_WIN_INSTANCE;
+            req->instance = wine_server_client_ptr( (void *)newval );
+            break;
+        case GWLP_WNDPROC:
+            req->flags = SET_WIN_UNICODE;
+            req->is_unicode = (win->flags & WIN_ISUNICODE) != 0;
+            break;
+        case GWLP_USERDATA:
+            if (size == sizeof(WORD)) newval = MAKELONG( newval, win->userdata >> 16 );
+            req->flags = SET_WIN_USERDATA;
+            req->user_data = newval;
+            break;
+        default:
+            req->flags = SET_WIN_EXTRA;
+            req->extra_offset = offset;
+            req->extra_size = size;
+            set_win_data( &req->extra_value, newval, size );
+        }
         if ((ok = !wine_server_call_err( req )))
         {
-            switch (offset)
+            switch(offset)
             {
             case GWL_STYLE:
                 win->dwStyle = newval;
                 win->dwExStyle = fix_exstyle(win->dwStyle, win->dwExStyle);
-                retval = reply->old_info;
+                retval = reply->old_style;
                 break;
             case GWL_EXSTYLE:
                 win->dwExStyle = newval;
-                retval = reply->old_info;
+                retval = reply->old_ex_style;
                 break;
             case GWLP_ID:
                 win->wIDmenu = newval;
-                retval = reply->old_info;
+                retval = reply->old_id;
                 break;
             case GWLP_HINSTANCE:
                 win->hInstance = (HINSTANCE)newval;
-                retval = reply->old_info;
+                retval = (ULONG_PTR)wine_server_get_ptr( reply->old_instance );
                 break;
             case GWLP_WNDPROC:
                 break;
             case GWLP_USERDATA:
                 win->userdata = newval;
-                retval = reply->old_info;
+                retval = reply->old_user_data;
                 break;
             default:
+                retval = get_win_data( (char *)win->wExtra + offset, size );
                 set_win_data( (char *)win->wExtra + offset, newval, size );
-                retval = reply->old_info;
                 break;
             }
         }
@@ -1589,25 +1480,26 @@
     return set_window_long( hwnd, offset, sizeof(LONG_PTR), newval, ansi );
 }
 
-BOOL set_window_pixel_format( HWND hwnd, int format, BOOL internal )
+BOOL win32u_set_window_pixel_format( HWND hwnd, int format, BOOL internal )
 {
     WND *win = get_win_ptr( hwnd );
 
     if (!win || win == WND_DESKTOP || win == WND_OTHER_PROCESS)
     {
-        WARN( "setting format %d on win %p semi-stub!\n", format, hwnd );
-        NtUserPostMessage( hwnd, WM_WINE_SETPIXELFORMAT, format, internal );
+        WARN( "setting format %d on win %p not supported\n", format, hwnd );
         return FALSE;
     }
-    if (!internal) win->pixel_format = format;
-    if (format) win->clip_clients = TRUE;
+    if (internal)
+        win->internal_pixel_format = format;
+    else
+        win->pixel_format = format;
     release_win_ptr( win );
 
     update_window_state( hwnd );
     return TRUE;
 }
 
-int get_window_pixel_format( HWND hwnd )
+int win32u_get_window_pixel_format( HWND hwnd )
 {
     WND *win = get_win_ptr( hwnd );
     int ret;
@@ -1615,7 +1507,7 @@
     if (!win || win == WND_DESKTOP || win == WND_OTHER_PROCESS)
     {
         WARN( "getting format on win %p not supported\n", hwnd );
-        return -1;
+        return 0;
     }
 
     ret = win->pixel_format;
@@ -1624,6 +1516,18 @@
     return ret;
 }
 
+static int window_has_client_surface( HWND hwnd, BOOL opengl_only )
+{
+    WND *win = get_win_ptr( hwnd );
+    BOOL ret;
+
+    if (!win || win == WND_DESKTOP || win == WND_OTHER_PROCESS) return FALSE;
+    ret = win->pixel_format || win->internal_pixel_format || (!list_empty(&win->vulkan_surfaces) && !opengl_only);
+    release_win_ptr( win );
+
+    return ret;
+}
+
 /***********************************************************************
  *           NtUserGetProp   (win32u.@)
  *
@@ -2001,7 +1905,6 @@
     UINT dpi = get_dpi_for_window( hwnd ), style_mask, ex_style_mask;
     RECT visible_rect, rect = {0};
 
-    if (get_present_rect( hwnd, &rect, get_thread_dpi() )) return rect;
     if (IsRectEmpty( &rects->window ) || EqualRect( &rects->window, &rects->client ) || shaped || !decorated_mode) return rects->window;
     if (!user_driver->pGetWindowStyleMasks( hwnd, style, ex_style, &style_mask, &ex_style_mask )) return rects->window;
     if (!NtUserAdjustWindowRect( &rect, style & style_mask, FALSE, ex_style & ex_style_mask, dpi )) return rects->window;
@@ -2019,6 +1922,21 @@
     return visible_rect;
 }
 
+static int force_present_to_surface(void)
+{
+    static int cached = -1;
+
+    if (cached == -1)
+    {
+        const char *sgi = getenv( "SteamGameId" );
+
+        cached = sgi &&
+                 (
+                    !strcmp(sgi, "803600")
+                 );
+    }
+    return cached;
+}
 
 static BOOL get_surface_rect( const RECT *visible_rect, RECT *surface_rect )
 {
@@ -2029,6 +1947,7 @@
     /* crop surfaces which are larger than the virtual screen rect, some applications create huge windows */
     if ((surface_rect->right - surface_rect->left > virtual_rect.right - virtual_rect.left ||
          surface_rect->bottom - surface_rect->top > virtual_rect.bottom - virtual_rect.top) &&
+        !force_present_to_surface() &&
         !intersect_rect( surface_rect, surface_rect, &virtual_rect ))
         return FALSE;
     OffsetRect( surface_rect, -visible_rect->left, -visible_rect->top );
@@ -2073,26 +1992,28 @@
 static struct window_surface *get_window_surface( HWND hwnd, UINT swp_flags, BOOL create_layered,
                                                   struct window_rects *rects, RECT *surface_rect )
 {
-    BOOL shaped, needs_surface, create_opaque, is_layered, is_child;
+    BOOL shaped, needs_surface, create_opaque, is_layered, is_child, is_opengl;
     HWND parent = NtUserGetAncestor( hwnd, GA_PARENT );
     struct window_surface *new_surface;
     struct window_rects monitor_rects;
-    UINT raw_dpi, style, ex_style;
+    UINT raw_dpi_num, raw_dpi_den, style, ex_style;
+    DWORD layered_flags;
     RECT dummy;
     HRGN shape;
 
     style = NtUserGetWindowLongW( hwnd, GWL_STYLE );
     ex_style = NtUserGetWindowLongW( hwnd, GWL_EXSTYLE );
     is_child = parent && parent != NtUserGetDesktopWindow();
+    is_opengl = window_has_client_surface( hwnd, TRUE );
 
-    if (is_child) get_win_monitor_dpi( parent, &raw_dpi );
-    else monitor_dpi_from_rect( rects->window, get_thread_dpi(), &raw_dpi );
+    if (is_child) get_win_monitor_dpi( parent, &raw_dpi_num, &raw_dpi_den );
+    else monitor_dpi_from_rect( rects->window, get_thread_dpi(), &raw_dpi_num, &raw_dpi_den );
 
     if (get_window_region( hwnd, FALSE, &shape, &dummy )) shaped = FALSE;
     else if ((shaped = !!shape)) NtGdiDeleteObjectApp( shape );
 
-    if (!get_present_rect( hwnd, &rects->visible, get_thread_dpi() )) rects->visible = rects->window;
-    if (is_child) monitor_rects = map_dpi_window_rects( *rects, get_thread_dpi(), raw_dpi );
+    rects->visible = rects->window;
+    if (is_child) monitor_rects = map_dpi_window_rects( *rects, get_thread_dpi() * raw_dpi_den, raw_dpi_num );
     else monitor_rects = map_window_rects_virt_to_raw( *rects, get_thread_dpi() );
 
     if (!user_driver->pWindowPosChanging( hwnd, swp_flags, shaped, &monitor_rects )) needs_surface = FALSE;
@@ -2118,15 +2039,21 @@
     if (IsRectEmpty( surface_rect )) needs_surface = FALSE;
     else if (create_layered || is_layered) needs_surface = TRUE;
 
-    if (needs_surface)
-        create_window_surface( hwnd, create_layered, surface_rect, raw_dpi, &new_surface );
+    if (is_opengl && !is_layered && !create_layered
+        && !(!create_opaque && NtUserGetLayeredWindowAttributes( hwnd, NULL, NULL, &layered_flags ) && layered_flags & LWA_COLORKEY))
+    {
+        if (new_surface) window_surface_release( new_surface );
+        new_surface = NULL;
+    }
+    else if (needs_surface)
+        create_window_surface( hwnd, create_layered, surface_rect, raw_dpi_num, raw_dpi_den, &new_surface );
     else if (new_surface && new_surface != &dummy_surface)
     {
         window_surface_release( new_surface );
         window_surface_add_ref( (new_surface = &dummy_surface) );
     }
 
-    if (new_surface && new_surface != &dummy_surface && !is_layered)
+    if (new_surface && !is_layered)
     {
         DWORD lwa_flags = 0, alpha_bits = -1;
         COLORREF key;
@@ -2141,28 +2068,20 @@
     return new_surface;
 }
 
-static HICON get_icon_info( HICON icon, ICONINFO *ii )
+static void update_children_window_state( HWND hwnd )
 {
-    return icon && NtUserGetIconInfo( icon, ii, NULL, NULL, NULL, 0 ) ? icon : NULL;
-}
+    HWND *children;
+    int i;
 
-HICON get_window_icon_info( HWND hwnd, UINT type, HICON icon, ICONINFO *ret )
-{
-    if ((icon = get_icon_info( icon, ret ))) return icon;
-    if ((icon = get_icon_info( (HICON)send_message( hwnd, WM_GETICON, type, 0 ), ret ))) return icon;
-    if ((icon = get_icon_info( (HICON)get_class_long_ptr( hwnd, GCLP_HICON, FALSE ), ret ))) return icon;
-    if (type == ICON_BIG)
+    if (!(children = list_window_children( hwnd ))) return;
+
+    for (i = 0; children[i]; i++)
     {
-        icon = LoadImageW( 0, (const WCHAR *)IDI_WINLOGO, IMAGE_ICON, 0, 0, LR_SHARED | LR_DEFAULTSIZE );
-        return get_icon_info( icon, ret );
+        if (!window_has_client_surface( children[i], FALSE )) continue;
+        update_window_state( children[i] );
     }
-    return NULL;
-}
 
-static BOOL is_fullscreen( const MONITORINFO *info, const RECT *rect )
-{
-    return rect->left <= info->rcMonitor.left && rect->right >= info->rcMonitor.right &&
-           rect->top <= info->rcMonitor.top && rect->bottom >= info->rcMonitor.bottom;
+    free( children );
 }
 
 /***********************************************************************
@@ -2175,23 +2094,21 @@
 {
     struct window_rects monitor_rects;
     WND *win;
-    HWND owner_hint, surface_win = 0, toplevel;
-    UINT raw_dpi, monitor_dpi, dpi = get_thread_dpi();
-    BOOL ret, is_layered, is_child, need_icons = FALSE;
+    HWND owner_hint, surface_win = 0, parent = NtUserGetAncestor( hwnd, GA_PARENT );
+    BOOL ret, is_fullscreen, is_layered, is_child;
     struct window_rects old_rects;
     RECT extra_rects[3];
     struct window_surface *old_surface;
-    HICON icon, icon_small;
-    ICONINFO ii, ii_small;
+    UINT raw_dpi_num, raw_dpi_den, monitor_dpi;
 
-    toplevel = NtUserGetAncestor( hwnd, GA_ROOT );
     is_layered = new_surface && new_surface->alpha_mask;
-    is_child = toplevel && toplevel != hwnd;
+    is_fullscreen = is_window_rect_full_screen( &new_rects->visible, get_thread_dpi() );
+    is_child = parent && parent != NtUserGetDesktopWindow();
 
-    if (is_child) monitor_dpi = get_win_monitor_dpi( toplevel, &raw_dpi );
-    else monitor_dpi = monitor_dpi_from_rect( new_rects->window, dpi, &raw_dpi );
+    if (is_child) monitor_dpi = get_win_monitor_dpi( parent, &raw_dpi_num, &raw_dpi_den );
+    else monitor_dpi = monitor_dpi_from_rect( new_rects->window, get_thread_dpi(), &raw_dpi_num, &raw_dpi_den );
 
-    get_window_rects( hwnd, COORDS_PARENT, &old_rects, dpi );
+    get_window_rects( hwnd, COORDS_PARENT, &old_rects, get_thread_dpi() );
     if (IsRectEmpty( &valid_rects[0] ) || is_layered) valid_rects = NULL;
 
     if (!(win = get_win_ptr( hwnd )) || win == WND_DESKTOP || win == WND_OTHER_PROCESS) return FALSE;
@@ -2205,6 +2122,9 @@
         valid_rects = NULL;
     }
 
+    if (is_child) monitor_rects = map_dpi_window_rects( *new_rects, get_thread_dpi() * raw_dpi_den, raw_dpi_num );
+    else monitor_rects = map_window_rects_virt_to_raw( *new_rects, get_thread_dpi() );
+
     SERVER_START_REQ( set_window_pos )
     {
         req->handle        = wine_server_user_handle( hwnd );
@@ -2227,7 +2147,8 @@
         }
         if (new_surface) req->paint_flags |= SET_WINPOS_PAINT_SURFACE;
         if (is_layered) req->paint_flags |= SET_WINPOS_LAYERED_WINDOW;
-        if (win->clip_clients) req->paint_flags |= SET_WINPOS_PIXEL_FORMAT;
+        else if (win->pixel_format || win->internal_pixel_format)
+             req->paint_flags |= SET_WINPOS_PIXEL_FORMAT;
 
         if ((ret = !wine_server_call( req )))
         {
@@ -2239,7 +2160,7 @@
             if (get_window_long( win->parent, GWL_EXSTYLE ) & WS_EX_LAYOUTRTL)
             {
                 RECT client = {0};
-                get_client_rect_rel( win->parent, COORDS_CLIENT, &client, dpi );
+                get_client_rect_rel( win->parent, COORDS_CLIENT, &client, get_thread_dpi() );
                 mirror_rect( &client, &win->rects.window );
                 mirror_rect( &client, &win->rects.client );
                 mirror_rect( &client, &win->rects.visible );
@@ -2258,32 +2179,6 @@
         if (((swp_flags & SWP_AGG_NOPOSCHANGE) != SWP_AGG_NOPOSCHANGE) ||
             (swp_flags & (SWP_HIDEWINDOW | SWP_SHOWWINDOW | SWP_STATECHANGED | SWP_FRAMECHANGED)))
             invalidate_dce( win, &old_rects.window );
-
-        if (win->dwStyle & WS_VISIBLE && !is_child && !win->has_icons)
-        {
-            icon = win->hIcon;
-            icon_small = win->hIconSmall2 ? win->hIconSmall2 : win->hIconSmall;
-            win->has_icons = need_icons = TRUE;
-        }
-
-        if (win->dwStyle & WS_THICKFRAME) swp_flags |= WINE_SWP_RESIZABLE;
-        if (is_child) monitor_rects = map_dpi_window_rects( *new_rects, dpi, raw_dpi );
-        else if (!IsRectEmpty( &win->present_rect ))
-        {
-            MONITORINFO monitor_info = monitor_info_from_rect( new_rects->window, dpi );
-            struct window_rects rects = { monitor_info.rcMonitor, monitor_info.rcMonitor, monitor_info.rcMonitor };
-            OffsetRect( &rects.client, -rects.client.left, -rects.client.top );
-            swp_flags |= WINE_SWP_FULLSCREEN;
-            swp_flags &= ~WINE_SWP_RESIZABLE;
-            monitor_rects = map_window_rects_virt_to_raw( rects, dpi );
-        }
-        else
-        {
-            MONITORINFO monitor_info = monitor_info_from_rect( new_rects->window, dpi );
-            if (is_fullscreen( &monitor_info, &new_rects->visible )) swp_flags |= WINE_SWP_FULLSCREEN;
-            if (is_fullscreen( &monitor_info, &new_rects->window )) swp_flags &= ~WINE_SWP_RESIZABLE;
-            monitor_rects = map_window_rects_virt_to_raw( *new_rects, dpi );
-        }
     }
 
     release_win_ptr( win );
@@ -2340,30 +2235,46 @@
             }
         }
 
-        if (need_icons && (icon = get_window_icon_info( hwnd, ICON_BIG, icon, &ii )))
-        {
-            icon_small = get_window_icon_info( hwnd, ICON_SMALL, icon_small, &ii_small );
-            user_driver->pSetWindowIcons( hwnd, icon, &ii, icon_small, &ii_small );
-        }
-
         owner_hint = NtUserGetWindowRelative(hwnd, GW_OWNER);
         /* fallback to any window that is right below our top left corner */
         if (!owner_hint) owner_hint = NtUserWindowFromPoint(new_rects->window.left - 1, new_rects->window.top - 1);
         if (owner_hint) owner_hint = NtUserGetAncestor(owner_hint, GA_ROOT);
 
-        user_driver->pWindowPosChanged( hwnd, insert_after, owner_hint, swp_flags, &monitor_rects,
-                                        get_driver_window_surface( new_surface, raw_dpi ) );
-        update_client_surfaces( toplevel );
+        user_driver->pWindowPosChanged( hwnd, insert_after, owner_hint, swp_flags, is_fullscreen, &monitor_rects,
+                                        get_driver_window_surface( new_surface, raw_dpi_num, raw_dpi_den ) );
+
+        update_children_window_state( hwnd );
+        vulkan_update_surfaces( hwnd );
     }
 
     return ret;
 }
 
+static HRGN expose_window_surface_rect( struct window_surface *surface, UINT flags, RECT dirty )
+{
+    HRGN region;
+
+    intersect_rect( &dirty, &dirty, &surface->rect );
+    add_bounds_rect( &surface->bounds, &dirty );
+
+    if (!surface->clip_region || !flags) return 0;
+
+    region = NtGdiCreateRectRgn( dirty.left, dirty.top, dirty.right, dirty.bottom );
+    if (NtGdiCombineRgn( region, region, surface->clip_region, RGN_DIFF ) <= NULLREGION)
+    {
+        NtGdiDeleteObjectApp( region );
+        region = 0;
+    }
+    return region;
+}
+
 static BOOL expose_window_surface( HWND hwnd, UINT flags, const RECT *rect, UINT dpi )
 {
     struct window_surface *surface;
     struct window_rects rects;
-    RECT exposed_rect;
+    RECT window_rect;
+    HRGN region = 0;
+    UINT win_dpi;
     WND *win;
 
     if (!(win = get_win_ptr( hwnd )) || win == WND_DESKTOP || win == WND_OTHER_PROCESS) return FALSE;
@@ -2371,26 +2282,37 @@
     rects = win->rects;
     release_win_ptr( win );
 
-    if (rect) exposed_rect = map_dpi_rect( *rect, dpi, get_dpi_for_window( hwnd ) );
-
-    if (!surface || surface == &dummy_surface)
+    if (rect)
     {
-        NtUserRedrawWindow( hwnd, rect ? &exposed_rect : NULL, NULL, flags );
-        if (surface) window_surface_release( surface );
-        return TRUE;
+        if ((win_dpi = get_dpi_for_window( hwnd )) != dpi)
+        {
+            window_rect = map_dpi_rect( *rect, dpi, win_dpi );
+            InflateRect( &window_rect, 1, 1 ); /* compensate rounding errors */
+        }
+        else window_rect = *rect;
     }
 
-    window_surface_lock( surface );
-    if (!rect) add_bounds_rect( &surface->bounds, &surface->rect );
-    else
+    if (surface)
     {
-        OffsetRect( &exposed_rect, rects.client.left - rects.visible.left, rects.client.top - rects.visible.top );
-        intersect_rect( &exposed_rect, &exposed_rect, &surface->rect );
-        add_bounds_rect( &surface->bounds, &exposed_rect );
+        window_surface_lock( surface );
+
+        if (!rect) add_bounds_rect( &surface->bounds, &surface->rect );
+        else
+        {
+            RECT dirty = window_rect;
+            OffsetRect( &dirty, rects.client.left - rects.visible.left, rects.client.top - rects.visible.top );
+            if (!(region = expose_window_surface_rect( surface, flags, dirty ))) flags = 0;
+            else NtGdiOffsetRgn( region, rects.client.left - rects.visible.left, rects.client.top - rects.visible.top );
+        }
+
+        window_surface_unlock( surface );
+        if (surface->alpha_mask) window_surface_flush( surface );
+
+        window_surface_release( surface );
     }
-    window_surface_unlock( surface );
-    if (surface->alpha_mask) window_surface_flush( surface );
-    window_surface_release( surface );
+
+    if (flags) NtUserRedrawWindow( hwnd, rect ? &window_rect : NULL, region, flags );
+    if (region) NtGdiDeleteObjectApp( region );
     return TRUE;
 }
 
@@ -2467,13 +2389,13 @@
     {
         UINT swp_flags = SWP_NOSIZE | SWP_NOMOVE | SWP_NOZORDER | SWP_NOACTIVATE | SWP_FRAMECHANGED |
             SWP_NOCLIENTSIZE | SWP_NOCLIENTMOVE;
-        UINT raw_dpi;
+        UINT raw_dpi_num, raw_dpi_den;
         HRGN monitor_hrgn;
 
         if (!redraw) swp_flags |= SWP_NOREDRAW;
 
-        get_win_monitor_dpi( hwnd, &raw_dpi );
-        monitor_hrgn = map_dpi_region( hrgn, get_thread_dpi(), raw_dpi );
+        get_win_monitor_dpi( hwnd, &raw_dpi_num, &raw_dpi_den );
+        monitor_hrgn = map_dpi_region( hrgn, get_thread_dpi() * raw_dpi_den, raw_dpi_num );
         user_driver->pSetWindowRgn( hwnd, monitor_hrgn, redraw );
         if (monitor_hrgn) NtGdiDeleteObjectApp( monitor_hrgn );
 
@@ -2523,7 +2445,7 @@
 {
     BOOL ret;
 
-    TRACE( "(%p,%s,%d,%x)\n", hwnd, debugstr_color(key), alpha, flags );
+    TRACE( "(%p,%s,%d,%x)\n", hwnd, debugstr_color(key), alpha, (int)flags );
 
     SERVER_START_REQ( set_window_layered_info )
     {
@@ -2552,7 +2474,6 @@
                                        const BLENDFUNCTION *blend, DWORD flags, const RECT *dirty )
 {
     DWORD swp_flags = SWP_NOSIZE | SWP_NOMOVE | SWP_NOZORDER | SWP_NOACTIVATE | SWP_NOREDRAW;
-    BYTE source_alpha = blend ? blend->SourceConstantAlpha : 0xff;
     struct window_rects new_rects;
     struct window_surface *surface;
     RECT surface_rect;
@@ -2607,11 +2528,7 @@
     apply_window_pos( hwnd, 0, swp_flags, surface, &new_rects, NULL );
     if (!surface) return FALSE;
 
-    if (!hdc_src || surface == &dummy_surface)
-    {
-        user_driver->pUpdateLayeredWindow( hwnd, source_alpha, flags );
-        ret = TRUE;
-    }
+    if (!hdc_src || surface == &dummy_surface) ret = TRUE;
     else
     {
         BLENDFUNCTION src_blend = { AC_SRC_OVER, 0, 255, 0 };
@@ -2632,6 +2549,7 @@
         if (pts_src) OffsetRect( &src_rect, pts_src->x, pts_src->y );
         NtGdiTransformPoints( hdc_src, (POINT *)&src_rect, (POINT *)&src_rect, 2, NtGdiDPtoLP );
 
+        if (flags & ULW_ALPHA) src_blend = *blend;
         ret = NtGdiAlphaBlend( hdc, rect.left, rect.top, rect.right - rect.left, rect.bottom - rect.top,
                                hdc_src, src_rect.left, src_rect.top, src_rect.right - src_rect.left, src_rect.bottom - src_rect.top,
                                *(DWORD *)&src_blend, 0 );
@@ -2643,7 +2561,7 @@
         if (!(flags & ULW_COLORKEY)) key = CLR_INVALID;
         window_surface_set_layered( surface, key, -1, 0xff000000 );
 
-        user_driver->pUpdateLayeredWindow( hwnd, source_alpha, flags );
+        user_driver->pUpdateLayeredWindow( hwnd, flags );
         window_surface_flush( surface );
     }
 
@@ -2705,10 +2623,10 @@
     int i, res;
     HWND ret, *list;
     POINT win_pt;
-    UINT dpi, raw_dpi;
+    UINT dpi;
 
     if (!hwnd) hwnd = get_desktop_window();
-    if (!(dpi = get_thread_dpi())) dpi = get_win_monitor_dpi( hwnd, &raw_dpi );
+    if (!(dpi = get_thread_dpi())) dpi = get_win_monitor_dpi( hwnd, NULL, NULL );
 
     *hittest = HTNOWHERE;
 
@@ -2718,7 +2636,7 @@
 
     for (i = 0; list[i]; i++)
     {
-        DWORD style = get_window_long( list[i], GWL_STYLE );
+        LONG style = get_window_long( list[i], GWL_STYLE );
 
         /* If window is minimized or disabled, return at once */
         if (style & WS_DISABLED)
@@ -2743,7 +2661,7 @@
     }
     ret = list[i];
     free( list );
-    TRACE( "scope %p (%d,%d) returning %p\n", hwnd, pt.x, pt.y, ret );
+    TRACE( "scope %p (%d,%d) returning %p\n", hwnd, (int)pt.x, (int)pt.y, ret );
     return ret;
 }
 
@@ -2778,7 +2696,7 @@
         if (!PtInRect( &rect, pt )) continue;
         if (flags & (CWP_SKIPINVISIBLE|CWP_SKIPDISABLED))
         {
-            DWORD style = get_window_long( list[i], GWL_STYLE );
+            LONG style = get_window_long( list[i], GWL_STYLE );
             if ((flags & CWP_SKIPINVISIBLE) && !(style & WS_VISIBLE)) continue;
             if ((flags & CWP_SKIPDISABLED) && (style & WS_DISABLED)) continue;
         }
@@ -2845,7 +2763,7 @@
     {
         if (!(wnd->dwStyle & WS_MINIMIZE))
         {
-            mon_info = monitor_info_from_window( wnd->handle, MONITOR_DEFAULTTOPRIMARY );
+            mon_info = monitor_info_from_window( wnd->obj.handle, MONITOR_DEFAULTTOPRIMARY );
             work_rect = mon_info.rcWork;
         }
 
@@ -3015,11 +2933,11 @@
     if (flags & PLACE_RECT) make_rect_onscreen( &wp.rcNormalPosition );
 
     TRACE( "%p: setting min %d,%d max %d,%d normal %s flags %x adjusted to min %d,%d max %d,%d normal %s\n",
-           hwnd, wndpl->ptMinPosition.x, wndpl->ptMinPosition.y,
-           wndpl->ptMaxPosition.x, wndpl->ptMaxPosition.y,
+           hwnd, (int)wndpl->ptMinPosition.x, (int)wndpl->ptMinPosition.y,
+           (int)wndpl->ptMaxPosition.x, (int)wndpl->ptMaxPosition.y,
            wine_dbgstr_rect(&wndpl->rcNormalPosition), flags,
-           wp.ptMinPosition.x, wp.ptMinPosition.y,
-           wp.ptMaxPosition.x, wp.ptMaxPosition.y,
+           (int)wp.ptMinPosition.x, (int)wp.ptMinPosition.y,
+           (int)wp.ptMaxPosition.x, (int)wp.ptMaxPosition.y,
            wine_dbgstr_rect(&wp.rcNormalPosition) );
 
     if (!win || win == WND_OTHER_PROCESS || win == WND_DESKTOP) return FALSE;
@@ -3122,10 +3040,11 @@
     user_handle_t *list;
     HWND retvalue = 0;
     int i = 0, size = 128, title_len;
+    ATOM atom = class ? get_int_atom_value( class ) : 0;
     NTSTATUS status;
 
     /* empty class is not the same as NULL class */
-    if (class && !class->Length && !IS_INTRESOURCE(class->Buffer)) return 0;
+    if (!atom && class && !class->Length) return 0;
 
     if (parent == HWND_MESSAGE) parent = get_hwnd_message_parent();
 
@@ -3137,7 +3056,8 @@
         {
             req->parent = wine_server_user_handle( parent );
             req->child  = wine_server_user_handle( child );
-            if (class) req->atom = wine_server_add_atom( req, class );
+            req->atom   = atom;
+            if (!atom && class) wine_server_add_data( req, class->Buffer, class->Length );
             wine_server_set_reply( req, list, size * sizeof(user_handle_t) );
             status = wine_server_call( req );
             size = reply->count;
@@ -3243,7 +3163,7 @@
         if (win == WND_OTHER_PROCESS) goto other_process;
         if (win != WND_DESKTOP)
         {
-            UINT raw_dpi, dpi_from = dpi ? dpi : get_win_monitor_dpi( hwnd_from, &raw_dpi );
+            UINT dpi_from = dpi ? dpi : get_win_monitor_dpi( hwnd_from, NULL, NULL );
             if (win->dwExStyle & WS_EX_LAYOUTRTL)
             {
                 mirror_from = TRUE;
@@ -3280,7 +3200,7 @@
         if (win == WND_OTHER_PROCESS) goto other_process;
         if (win != WND_DESKTOP)
         {
-            UINT raw_dpi, dpi_to = dpi ? dpi : get_win_monitor_dpi( hwnd_to, &raw_dpi );
+            UINT dpi_to = dpi ? dpi : get_win_monitor_dpi( hwnd_to, NULL, NULL );
             POINT pt = { 0, 0 };
             if (win->dwExStyle & WS_EX_LAYOUTRTL)
             {
@@ -3467,9 +3387,9 @@
 static void map_dpi_winpos( WINDOWPOS *winpos )
 {
     RECT rect = {winpos->x, winpos->y, winpos->x + winpos->cx, winpos->y + winpos->cy};
-    UINT raw_dpi, dpi_from = get_thread_dpi(), dpi_to = get_dpi_for_window( winpos->hwnd );
+    UINT dpi_from = get_thread_dpi(), dpi_to = get_dpi_for_window( winpos->hwnd );
 
-    if (!dpi_from) dpi_from = get_win_monitor_dpi( winpos->hwnd, &raw_dpi );
+    if (!dpi_from) dpi_from = get_win_monitor_dpi( winpos->hwnd, NULL, NULL );
     rect = map_dpi_rect( rect, dpi_from, dpi_to );
     winpos->x = rect.left;
     winpos->y = rect.top;
@@ -3506,6 +3426,32 @@
         }
         else
         {
+            const char *sgi = getenv( "SteamGameId" );
+
+            /* CW-Bug-Id: #25036 Escape from Monkey Island™ (730830) Alt+Tab shifts game downward
+             * when opened via config launcher.
+             *
+             * The game calls ShowWindow(hwnd, SW_RESTORE) whenever WM_ACTIVATE or WM_SYSCOMMAND
+             * SC_RESTORE messages are received. And the game is created with a size of 648x514.
+             * The multiple ShowWindow(hwnd, SW_RESTORE) calls will restore its exclusive
+             * fullscreen from 640x480 to 648x514. However, moving or resizing exclusive fullscreen
+             * windows on Windows are not allowed. Due to the missing support for exclusive
+             * fullscreen on Wine, work around it by forcing fullscreen here.
+             */
+            if (sgi && (!strcmp( sgi, "730830" ))
+                && !(win->dwStyle & WS_MINIMIZE) && win->parent == get_desktop_window()
+                && is_window_rect_full_screen( &old_rects->visible, get_thread_dpi() ))
+            {
+                RECT rect = {winpos->x, winpos->y,
+                             winpos->x + new_rects->window.right - new_rects->window.left,
+                             winpos->y + new_rects->window.bottom - new_rects->window.top};
+                MONITORINFO info = monitor_info_from_rect( rect, get_thread_dpi() );
+                WARN( "Forcing fullscreen window %p size from %s to %s\n", winpos->hwnd,
+                      wine_dbgstr_rect(&rect), wine_dbgstr_rect(&info.rcMonitor) );
+                winpos->cx = info.rcMonitor.right - info.rcMonitor.left;
+                winpos->cy = info.rcMonitor.bottom - info.rcMonitor.top;
+            }
+
             new_rects->window.right  = new_rects->window.left + winpos->cx;
             new_rects->window.bottom = new_rects->window.top + winpos->cy;
         }
@@ -3521,6 +3467,26 @@
             winpos->x = -32000;
             winpos->y = -32000;
         }
+        else
+        {
+            const char *sgi = getenv( "SteamGameId" );
+
+            /* CW-Bug-Id: #24943 Space Engineers (244850) has issues with fullscreen focus loss with
+             * non-native aspect ratio */
+            if (sgi && !strcmp( sgi, "244850" )
+                && !(win->dwStyle & WS_MINIMIZE) && win->parent == get_desktop_window()
+                && is_window_rect_full_screen( &old_rects->visible, get_thread_dpi() ))
+            {
+                RECT rect = {winpos->x, winpos->y,
+                             winpos->x + new_rects->window.right - new_rects->window.left,
+                             winpos->y + new_rects->window.bottom - new_rects->window.top};
+                MONITORINFO info = monitor_info_from_rect( rect, get_thread_dpi() );
+                WARN( "Forcing fullscreen window %p origin from %s to %s\n", winpos->hwnd,
+                      wine_dbgstr_rect(&rect), wine_dbgstr_rect(&info.rcMonitor) );
+                winpos->x = info.rcMonitor.left;
+                winpos->y = info.rcMonitor.top;
+            }
+        }
         new_rects->window.left    = winpos->x;
         new_rects->window.top     = winpos->y;
         new_rects->window.right  += winpos->x - old_rects->window.left;
@@ -3694,7 +3660,7 @@
         RtlSetLastWin32Error( ERROR_INVALID_WINDOW_HANDLE );
         return FALSE;
     }
-    winpos->hwnd = win->handle;  /* make it a full handle */
+    winpos->hwnd = win->obj.handle;  /* make it a full handle */
 
     /* Finally make sure that all coordinates are valid */
     if (winpos->x < -32768) winpos->x = -32768;
@@ -3950,7 +3916,7 @@
         if ((style & (WS_CHILD | WS_POPUP)) == WS_CHILD)
             send_message( winpos->hwnd, WM_CHILDACTIVATE, 0, 0 );
         else if (!(style & WS_MINIMIZE))
-            set_foreground_window( winpos->hwnd, FALSE, FALSE );
+            set_foreground_window( winpos->hwnd, FALSE );
     }
 
     if(!(orig_flags & SWP_DEFERERASE))
@@ -4037,16 +4003,15 @@
 
 typedef struct
 {
+    struct user_object obj;
     INT        count;
     INT        suggested_count;
     HWND       parent;
     WINDOWPOS *winpos;
 } DWP;
 
-/***********************************************************************
- *           NtUserBeginDeferWindowPos (win32u.@)
- */
-HDWP WINAPI NtUserBeginDeferWindowPos( INT count )
+/* see BeginDeferWindowPos */
+HDWP begin_defer_window_pos( INT count )
 {
     HDWP handle = 0;
     DWP *dwp;
@@ -4068,7 +4033,7 @@
     dwp->suggested_count = count;
 
     if (!(dwp->winpos = malloc( count * sizeof(WINDOWPOS) )) ||
-        !(handle = alloc_user_handle( dwp, NTUSER_OBJ_WINPOS )))
+        !(handle = alloc_user_handle( &dwp->obj, NTUSER_OBJ_WINPOS )))
     {
         free( dwp->winpos );
         free( dwp );
@@ -4246,13 +4211,12 @@
  */
 static BOOL can_activate_window( HWND hwnd )
 {
-    DWORD style;
+    LONG style;
 
     if (!hwnd) return FALSE;
     style = get_window_long( hwnd, GWL_STYLE );
     if (!(style & WS_VISIBLE)) return FALSE;
     if ((style & (WS_POPUP|WS_CHILD)) == WS_CHILD) return FALSE;
-    if (style & WS_MINIMIZE) return FALSE;
     return !(style & WS_DISABLED);
 }
 
@@ -4296,7 +4260,7 @@
     TRACE( "win = %p fg = %p\n", hwnd_to, fg );
     if (!fg || hwnd == fg)
     {
-        if (set_foreground_window( hwnd_to, FALSE, FALSE )) return;
+        if (set_foreground_window( hwnd_to, FALSE )) return;
     }
     if (NtUserSetActiveWindow( hwnd_to )) NtUserSetActiveWindow( 0 );
 }
@@ -4323,11 +4287,11 @@
  */
 MINMAXINFO get_min_max_info( HWND hwnd )
 {
-    DWORD style = get_window_long( hwnd, GWL_STYLE );
-    DWORD exstyle = get_window_long( hwnd, GWL_EXSTYLE );
+    LONG style = get_window_long( hwnd, GWL_STYLE );
+    LONG exstyle = get_window_long( hwnd, GWL_EXSTYLE );
     UINT context;
     RECT rc_max, rc_primary;
-    DWORD adjusted_style;
+    LONG adjusted_style;
     MINMAXINFO minmax;
     INT xinc, yinc;
     RECT rc;
@@ -4396,10 +4360,10 @@
     }
 
     TRACE( "%d %d / %d %d / %d %d / %d %d\n",
-           minmax.ptMaxSize.x, minmax.ptMaxSize.y,
-           minmax.ptMaxPosition.x, minmax.ptMaxPosition.y,
-           minmax.ptMaxTrackSize.x, minmax.ptMaxTrackSize.y,
-           minmax.ptMinTrackSize.x, minmax.ptMinTrackSize.y );
+           (int)minmax.ptMaxSize.x, (int)minmax.ptMaxSize.y,
+           (int)minmax.ptMaxPosition.x, (int)minmax.ptMaxPosition.y,
+           (int)minmax.ptMaxTrackSize.x, (int)minmax.ptMaxTrackSize.y,
+           (int)minmax.ptMinTrackSize.x, (int)minmax.ptMinTrackSize.y );
 
     minmax.ptMaxTrackSize.x = max( minmax.ptMaxTrackSize.x, minmax.ptMinTrackSize.x );
     minmax.ptMaxTrackSize.y = max( minmax.ptMaxTrackSize.y, minmax.ptMinTrackSize.y );
@@ -4544,7 +4508,7 @@
 static UINT window_min_maximize( HWND hwnd, UINT cmd, RECT *rect )
 {
     UINT swp_flags = 0;
-    DWORD old_style;
+    LONG old_style;
     MINMAXINFO minmax;
     WINDOWPLACEMENT wpl;
 
@@ -4592,7 +4556,7 @@
                 NtUserSetFocus( 0 );
         }
 
-        old_style = set_window_style_bits( hwnd, WS_MINIMIZE, WS_MAXIMIZE );
+        old_style = set_window_style( hwnd, WS_MINIMIZE, WS_MAXIMIZE );
 
         wpl.ptMinPosition = get_minimized_pos( hwnd, wpl.ptMinPosition );
 
@@ -4609,7 +4573,7 @@
 
         minmax = get_min_max_info( hwnd );
 
-        old_style = set_window_style_bits( hwnd, WS_MAXIMIZE, WS_MINIMIZE );
+        old_style = set_window_style( hwnd, WS_MAXIMIZE, WS_MINIMIZE );
         if (old_style & WS_MINIMIZE)
             win_set_flags( hwnd, WIN_RESTORE_MAX, 0 );
 
@@ -4625,14 +4589,14 @@
     case SW_SHOWNORMAL:
     case SW_RESTORE:
     case SW_SHOWDEFAULT: /* FIXME: should have its own handler */
-        old_style = set_window_style_bits( hwnd, 0, WS_MINIMIZE | WS_MAXIMIZE );
+        old_style = set_window_style( hwnd, 0, WS_MINIMIZE | WS_MAXIMIZE );
         if (old_style & WS_MINIMIZE)
         {
             if (win_get_flags( hwnd ) & WIN_RESTORE_MAX)
             {
                 /* Restore to maximized position */
                 minmax = get_min_max_info( hwnd );
-                set_window_style_bits( hwnd, WS_MAXIMIZE, 0 );
+                set_window_style( hwnd, WS_MAXIMIZE, 0 );
                 swp_flags |= SWP_STATECHANGED;
                 SetRect( rect, minmax.ptMaxPosition.x, minmax.ptMaxPosition.y,
                          minmax.ptMaxPosition.x + minmax.ptMaxSize.x,
@@ -4653,10 +4617,8 @@
     return swp_flags;
 }
 
-/***********************************************************************
- *           NtUserArrangeIconicWindows (win32u.@)
- */
-UINT WINAPI NtUserArrangeIconicWindows( HWND parent )
+/* see ArrangeIconicWindows */
+static UINT arrange_iconic_windows( HWND parent )
 {
     int width, height, count = 0;
     MINIMIZEDMETRICS metrics;
@@ -4707,6 +4669,7 @@
     struct window_surface *surface;
     struct window_rects new_rects;
 
+    if (!hwnd || hwnd == get_desktop_window()) return;
     if (!is_current_thread_window( hwnd ))
     {
         NtUserPostMessage( hwnd, WM_WINE_UPDATEWINDOWSTATE, 0, 0 );
@@ -4722,6 +4685,7 @@
     if (surface) window_surface_release( surface );
 
     set_thread_dpi_awareness_context( context );
+    vulkan_update_surfaces( hwnd );
 }
 
 /***********************************************************************
@@ -4731,26 +4695,37 @@
  */
 static BOOL show_window( HWND hwnd, INT cmd )
 {
+    static volatile LONG first_window = 1;
     WND *win;
     HWND parent;
-    DWORD style = get_window_long( hwnd, GWL_STYLE ), new_style;
+    LONG style = get_window_long( hwnd, GWL_STYLE ), new_style;
     BOOL was_visible = (style & WS_VISIBLE) != 0;
     BOOL show_flag = TRUE;
     RECT newPos = {0, 0, 0, 0};
     UINT new_swp, swp = 0, context;
+    const char *sgi;
 
     TRACE( "hwnd=%p, cmd=%d, was_visible %d\n", hwnd, cmd, was_visible );
 
+    /* Batman: Arkham City minimises its main window while the splash window is shown, but under gamescope, the
+     * game then restores the window and it becomes the foreground, resulting in broken splash display. The game
+     * later calls ShowWindow() with SW_SHOWNORMAL, so it's safe to suppress the restoration. CW-Bug-Id: #25955 */
+    if (cmd == SW_RESTORE
+            && user_driver->pHasWindowManager( "steamcompmgr" )
+            && (sgi = getenv( "SteamGameId" )) && !strcmp( sgi, "200260" ))
+        return was_visible;
+
     context = set_thread_dpi_awareness_context( get_window_dpi_awareness_context( hwnd ));
 
     if ((!(style & (WS_POPUP | WS_CHILD))
          || ((style & (WS_POPUP | WS_CHILD | WS_CAPTION)) == (WS_POPUP | WS_CAPTION)))
-        && !get_window_relative( hwnd, GW_OWNER )
-        && set_startup_info_flags( STARTF_USESHOWWINDOW, 0 ) & STARTF_USESHOWWINDOW)
+        && InterlockedExchange( &first_window, 0 ))
     {
-        if (cmd == SW_SHOW || cmd == SW_SHOWNORMAL || cmd == SW_SHOWDEFAULT)
+        RTL_USER_PROCESS_PARAMETERS *params = NtCurrentTeb()->Peb->ProcessParameters;
+
+        if (params->dwFlags & STARTF_USESHOWWINDOW && (cmd == SW_SHOW || cmd == SW_SHOWNORMAL || cmd == SW_SHOWDEFAULT))
         {
-            cmd = startup_show_window;
+            cmd = params->wShowWindow;
             TRACE( "hwnd=%p, using cmd %d from startup info.\n", hwnd, cmd );
         }
     }
@@ -4841,8 +4816,8 @@
     if (parent && !is_window_visible( parent ) && !(swp & SWP_STATECHANGED))
     {
         /* if parent is not visible simply toggle WS_VISIBLE and return */
-        if (show_flag) set_window_style_bits( hwnd, WS_VISIBLE, 0 );
-        else set_window_style_bits( hwnd, 0, WS_VISIBLE );
+        if (show_flag) set_window_style( hwnd, WS_VISIBLE, 0 );
+        else set_window_style( hwnd, 0, WS_VISIBLE );
     }
     else
         NtUserSetWindowPos( hwnd, HWND_TOP, newPos.left, newPos.top,
@@ -4965,10 +4940,8 @@
     return send_message( hwnd, WM_WINE_SHOWWINDOW, cmd, 0 );
 }
 
-/***********************************************************************
- *           NtUserShowOwnedPopups (win32u.@)
- */
-BOOL WINAPI NtUserShowOwnedPopups( HWND owner, BOOL show )
+/* see ShowOwnedPopups */
+BOOL show_owned_popups( HWND owner, BOOL show )
 {
     int count = 0;
     HWND *win_array = list_window_children( 0 );
@@ -5050,7 +5023,7 @@
 
         win = get_win_ptr( hwnd );
         if (!win || win == WND_OTHER_PROCESS || win == WND_DESKTOP) return FALSE;
-        hwnd = win->handle;  /* make it a full handle */
+        hwnd = win->obj.handle;  /* make it a full handle */
 
         if (info->dwFlags) wparam = !(win->flags & WIN_NCACTIVATED);
         else wparam = (hwnd == NtUserGetForegroundWindow());
@@ -5065,10 +5038,8 @@
     }
 }
 
-/***********************************************************************
- *           NtUserGetWindowContextHelpId   (win32u.@)
- */
-DWORD WINAPI NtUserGetWindowContextHelpId( HWND hwnd )
+/* see GetWindowContextHelpId */
+DWORD get_window_context_help_id( HWND hwnd )
 {
     DWORD retval;
     WND *win = get_win_ptr( hwnd );
@@ -5083,10 +5054,8 @@
     return retval;
 }
 
-/***********************************************************************
- *           NtUserSetWindowContextHelpId   (win32u.@)
- */
-BOOL WINAPI NtUserSetWindowContextHelpId( HWND hwnd, DWORD id )
+/* see SetWindowContextHelpId */
+static BOOL set_window_context_help_id( HWND hwnd, DWORD id )
 {
     WND *win = get_win_ptr( hwnd );
     if (!win || win == WND_DESKTOP) return FALSE;
@@ -5222,14 +5191,16 @@
  */
 LRESULT destroy_window( HWND hwnd )
 {
-    struct list drawables = LIST_INIT( drawables );
+    struct list vulkan_surfaces = LIST_INIT(vulkan_surfaces);
     struct window_surface *surface;
     HMENU menu = 0, sys_menu;
     WND *win;
-    HWND *children, toplevel = NtUserGetAncestor( hwnd, GA_ROOT );
+    HWND toplevel, *children;
 
     TRACE( "%p\n", hwnd );
 
+    toplevel = NtUserGetAncestor( hwnd, GA_ROOT );
+
     unregister_imm_window( hwnd );
 
     /* free child windows */
@@ -5258,7 +5229,7 @@
 
     send_message( hwnd, WM_NCDESTROY, 0, 0 );
 
-    if (toplevel && toplevel != hwnd) update_window_state( toplevel );
+    if (toplevel != hwnd) update_window_state( toplevel );
 
     /* FIXME: do we need to fake QS_MOUSEMOVE wakebit? */
 
@@ -5268,15 +5239,14 @@
     if ((win->dwStyle & (WS_CHILD | WS_POPUP)) != WS_CHILD)
         menu = (HMENU)win->wIDmenu;
     sys_menu = win->hSysMenu;
-    free_dce( win->dce, hwnd, &drawables );
+    free_dce( win->dce, hwnd );
     win->dce = NULL;
     NtUserDestroyCursor( win->hIconSmall2, 0 );
+    list_move_tail( &vulkan_surfaces, &win->vulkan_surfaces );
     surface = win->surface;
     win->surface = NULL;
     release_win_ptr( win );
 
-    release_opengl_drawables( &drawables );
-
     NtUserDestroyMenu( menu );
     NtUserDestroyMenu( sys_menu );
     if (surface)
@@ -5285,9 +5255,7 @@
         window_surface_release( surface );
     }
 
-    detach_client_surfaces( hwnd );
-    if (win->current_drawable) opengl_drawable_release( win->current_drawable );
-    if (win->unused_drawable) opengl_drawable_release( win->unused_drawable );
+    vulkan_detach_surfaces( &vulkan_surfaces );
     user_driver->pDestroyWindow( hwnd );
 
     free_window_handle( hwnd );
@@ -5387,46 +5355,17 @@
  */
 void destroy_thread_windows(void)
 {
-    struct destroy_entry
-    {
-        HWND handle;
-        HMENU menu;
-        HMENU sys_menu;
-        struct opengl_drawable *current_drawable;
-        struct opengl_drawable *unused_drawable;
-        struct window_surface *surface;
-        struct destroy_entry *next;
-    } *entry, *free_list = NULL;
-    struct list drawables = LIST_INIT(drawables);
+    WND *win, *free_list = NULL;
     HANDLE handle = 0;
-    WND *win;
-
-    /* recycle WND structs as destroy_entry structs */
-    C_ASSERT( sizeof(struct destroy_entry) <= sizeof(WND) );
 
     user_lock();
-    while ((win = next_thread_user_object( GetCurrentThreadId(), &handle, NTUSER_OBJ_WINDOW )))
+    while ((win = next_process_user_handle_ptr( &handle, NTUSER_OBJ_WINDOW )))
     {
-        BOOL is_child = (win->dwStyle & (WS_CHILD | WS_POPUP)) == WS_CHILD;
-        struct destroy_entry tmp = {0};
-
-        free_dce( win->dce, win->handle, &drawables );
+        if (win->tid != GetCurrentThreadId()) continue;
+        free_dce( win->dce, win->obj.handle );
         set_user_handle_ptr( handle, NULL );
-        free( win->pScroll );
-        free( win->text );
-
-        /* recycle the WND struct as a destroy_entry struct */
-        entry = (struct destroy_entry *)win;
-        tmp.handle = win->handle;
-        if (!is_child) tmp.menu = (HMENU)win->wIDmenu;
-        tmp.sys_menu = win->hSysMenu;
-        tmp.current_drawable = win->current_drawable;
-        tmp.unused_drawable = win->unused_drawable;
-        tmp.surface = win->surface;
-        *entry = tmp;
-
-        entry->next = free_list;
-        free_list = entry;
+        win->userdata = (UINT_PTR)free_list;
+        free_list = win;
     }
     if (free_list)
     {
@@ -5439,26 +5378,25 @@
     }
     user_unlock();
 
-    release_opengl_drawables( &drawables );
-
-    while ((entry = free_list))
+    while ((win = free_list))
     {
-        free_list = entry->next;
-        TRACE( "destroying %p\n", entry );
+        free_list = (WND *)win->userdata;
+        TRACE( "destroying %p\n", win );
 
-        detach_client_surfaces( entry->handle );
-        user_driver->pDestroyWindow( entry->handle );
-        if (entry->current_drawable) opengl_drawable_release( entry->current_drawable );
-        if (entry->unused_drawable) opengl_drawable_release( entry->unused_drawable );
+        user_driver->pDestroyWindow( win->obj.handle );
+        vulkan_detach_surfaces( &win->vulkan_surfaces );
 
-        NtUserDestroyMenu( entry->menu );
-        NtUserDestroyMenu( entry->sys_menu );
-        if (entry->surface)
+        if ((win->dwStyle & (WS_CHILD | WS_POPUP)) != WS_CHILD && win->wIDmenu)
+            NtUserDestroyMenu( UlongToHandle(win->wIDmenu) );
+        if (win->hSysMenu) NtUserDestroyMenu( win->hSysMenu );
+        if (win->surface)
         {
-            register_window_surface( entry->surface, NULL );
-            window_surface_release( entry->surface );
+            register_window_surface( win->surface, NULL );
+            window_surface_release( win->surface );
         }
-        free( entry );
+        free( win->pScroll );
+        free( win->text );
+        free( win );
     }
 }
 
@@ -5467,8 +5405,9 @@
  *
  * Create a window handle with the server.
  */
-static WND *create_window_handle( HWND parent, HWND owner, UNICODE_STRING *name, HINSTANCE class_instance,
-                                  HINSTANCE instance, BOOL ansi, DWORD style, DWORD ex_style )
+static WND *create_window_handle( HWND parent, HWND owner, UNICODE_STRING *name,
+                                  HINSTANCE instance, BOOL ansi,
+                                  DWORD style, DWORD ex_style )
 {
     UINT dpi_context = get_thread_dpi_awareness_context();
     HWND handle = 0, full_parent = 0, full_owner = 0;
@@ -5478,20 +5417,21 @@
 
     SERVER_START_REQ( create_window )
     {
-        req->parent          = wine_server_user_handle( parent );
-        req->owner           = wine_server_user_handle( owner );
-        req->class_instance  = wine_server_client_ptr( class_instance );
-        req->instance        = wine_server_client_ptr( instance );
-        req->dpi_context     = dpi_context;
-        req->style           = style;
-        req->ex_style        = ex_style;
-        req->atom            = wine_server_add_atom( req, name );
+        req->parent    = wine_server_user_handle( parent );
+        req->owner     = wine_server_user_handle( owner );
+        req->instance  = wine_server_client_ptr( instance );
+        req->dpi_context = dpi_context;
+        req->style     = style;
+        req->ex_style  = ex_style;
+        if (!(req->atom = get_int_atom_value( name )) && name->Length)
+            wine_server_add_data( req, name->Buffer, name->Length );
         if (!wine_server_call_err( req ))
         {
             handle      = wine_server_ptr_handle( reply->handle );
             full_parent = wine_server_ptr_handle( reply->parent );
             full_owner  = wine_server_ptr_handle( reply->owner );
             extra_bytes = reply->extra;
+            dpi_context = reply->dpi_context;
             class       = wine_server_get_ptr( reply->class_ptr );
         }
     }
@@ -5499,7 +5439,7 @@
 
     if (!handle)
     {
-        WARN( "error %d creating window\n", RtlGetLastWin32Error() );
+        WARN( "error %d creating window\n", (int)RtlGetLastWin32Error() );
         return NULL;
     }
 
@@ -5519,7 +5459,7 @@
     {
         struct ntuser_thread_info *thread_info = NtUserGetThreadInfo();
 
-        if (is_desktop_class( name ))
+        if (name->Buffer == (const WCHAR *)DESKTOP_CLASS_ATOM)
         {
             if (!thread_info->top_window) thread_info->top_window = HandleToUlong( full_parent ? full_parent : handle );
             else assert( full_parent == UlongToHandle( thread_info->top_window ));
@@ -5536,14 +5476,16 @@
 
     user_lock();
 
-    win->handle     = handle;
+    win->obj.handle = handle;
+    win->obj.type   = NTUSER_OBJ_WINDOW;
     win->parent     = full_parent;
     win->owner      = full_owner;
     win->class      = class;
     win->winproc    = get_class_winproc( class );
-    win->hInstance  = instance;
     win->cbWndExtra = extra_bytes;
-    set_user_handle_ptr( handle, win );
+    win->dpi_context = dpi_context;
+    list_init( &win->vulkan_surfaces );
+    set_user_handle_ptr( handle, &win->obj );
     if (is_winproc_unicode( win->winproc, !ansi )) win->flags |= WIN_ISUNICODE;
     return win;
 }
@@ -5617,7 +5559,7 @@
 
     if (!dpi_from || !dpi_to)
     {
-        UINT raw_dpi, mon_dpi = monitor_dpi_from_rect( rect, get_thread_dpi(), &raw_dpi );
+        UINT mon_dpi = monitor_dpi_from_rect( rect, get_thread_dpi(), NULL, NULL );
         if (!dpi_from) dpi_from = mon_dpi;
         else dpi_to = mon_dpi;
     }
@@ -5629,16 +5571,62 @@
     cs->cy = rect.bottom - rect.top;
 }
 
+int disable_gamescope_max_size_hack(void)
+{
+    static int cached = -1;
+
+    if (cached == -1)
+    {
+        const char *s = getenv( "WINE_DISABLE_GAMESCOPE_MAX_SIZE_HACK" );
+
+        cached = s && *s != '0';
+    }
+
+    return cached;
+}
+
+static void style_fixup_workarounds( CREATESTRUCTW *cs, unsigned int class_name_len )
+{
+    static const WCHAR BlockClickMASKWndW[] = {'B','l','o','c','k','C','l','i','c','k','M','A','S','K','W','n','d'};
+    static const struct
+    {
+        const WCHAR *class_name;
+        const char *wm;
+        const char *game_id;
+        unsigned exstyle;
+    }
+    workarounds[] =
+    {
+        { BlockClickMASKWndW, "steamcompmgr", "3839850", WS_EX_NOACTIVATE },
+    };
+    const char *sgi = getenv( "SteamGameId" );
+    unsigned int i;
+
+    if (cs->lpszClass == (LPCWSTR)DESKTOP_CLASS_ATOM) return;
+
+    for (i = 0; i < ARRAY_SIZE(workarounds); ++i)
+    {
+        if ((!workarounds[i].wm || user_driver->pHasWindowManager( workarounds[i].wm ))
+            && (!workarounds[i].game_id || (sgi && !strcmp( sgi, workarounds[i].game_id )))
+            && (!workarounds[i].class_name ||
+                 (class_name_len == wcslen( workarounds[i].class_name ) * sizeof(WCHAR)
+                 && !memcmp( cs->lpszClass, workarounds[i].class_name, class_name_len ))))
+        {
+            FIXME( "HACK: adding %#x exstyle for %s.\n", workarounds[i].exstyle, debugstr_w( cs->lpszClass ));
+            cs->dwExStyle |= workarounds[i].exstyle;
+        }
+    }
+}
+
 /***********************************************************************
  *           NtUserCreateWindowEx (win32u.@)
  */
 HWND WINAPI NtUserCreateWindowEx( DWORD ex_style, UNICODE_STRING *class_name,
                                   UNICODE_STRING *version, UNICODE_STRING *window_name,
                                   DWORD style, INT x, INT y, INT cx, INT cy,
-                                  HWND parent, HMENU menu, HINSTANCE class_instance, void *params,
-                                  DWORD flags, HINSTANCE instance, const WCHAR *class, BOOL ansi )
+                                  HWND parent, HMENU menu, HINSTANCE instance, void *params,
+                                  DWORD flags, HINSTANCE client_instance, DWORD unk, BOOL ansi )
 {
-    WCHAR base_nameW[MAX_ATOM_LEN + 1];
     UINT win_dpi, context;
     struct window_surface *surface;
     struct window_rects new_rects;
@@ -5649,23 +5637,23 @@
     RECT surface_rect;
     WND *win;
 
-    TRACE( "ex_style %#x, class_name %s, version %s, window_name %s, style %#x, x %u, y %u, cx %u, cy %u, "
-           "parent %p, menu %p, class_instance %p, params %p, flags %#x, instance %p, class %s, ansi %u\n",
-           ex_style, debugstr_us(class_name), debugstr_us(version), debugstr_us(window_name), style, x, y, cx, cy,
-           parent, menu, class_instance, params, flags, instance, debugstr_w(class), ansi );
+    static const WCHAR messageW[] = {'M','e','s','s','a','g','e'};
 
     cs.lpCreateParams = params;
-    cs.hInstance  = instance ? instance : class_instance;
+    cs.hInstance  = client_instance ? client_instance : instance;
     cs.hMenu      = menu;
     cs.hwndParent = parent;
     cs.style      = style;
     cs.dwExStyle  = ex_style;
     cs.lpszName   = window_name ? window_name->Buffer : NULL;
+    cs.lpszClass  = class_name->Buffer;
     cs.x  = x;
     cs.y  = y;
     cs.cx = cx;
     cs.cy = cy;
 
+    style_fixup_workarounds( &cs, class_name->Length );
+
     /* Find the parent window */
     if (parent == HWND_MESSAGE)
     {
@@ -5695,7 +5683,9 @@
         }
 
         /* are we creating the desktop or HWND_MESSAGE parent itself? */
-        if (!is_desktop_class( class_name ) && !is_message_class( class_name ))
+        if (class_name->Buffer != (LPCWSTR)DESKTOP_CLASS_ATOM &&
+            (class_name->Length != sizeof(messageW) ||
+             wcsnicmp( class_name->Buffer, messageW, ARRAYSIZE(messageW) )))
         {
             if (get_process_layout() & LAYOUT_RTL) cs.dwExStyle |= WS_EX_LAYOUTRTL;
             parent = get_desktop_window();
@@ -5709,13 +5699,14 @@
 
     style = cs.style & ~WS_VISIBLE;
     ex_style = cs.dwExStyle & ~WS_EX_LAYERED;
-    if (!(win = create_window_handle( parent, owner, class_name, class_instance,
-                                      cs.hInstance, ansi, style, ex_style )))
+    if (!(win = create_window_handle( parent, owner, class_name, instance, ansi, style, ex_style )))
         return 0;
-    hwnd = win->handle;
+    hwnd = win->obj.handle;
 
     /* Fill the window structure */
 
+    win->tid         = GetCurrentThreadId();
+    win->hInstance   = cs.hInstance;
     win->text        = NULL;
     win->dwStyle     = style;
     win->dwExStyle   = ex_style;
@@ -5727,7 +5718,6 @@
     win->hIconSmall  = 0;
     win->hIconSmall2 = 0;
     win->hSysMenu    = 0;
-    win->swap_interval = 1;
 
     win->min_pos.x = win->min_pos.y = -1;
     win->max_pos.x = win->max_pos.y = -1;
@@ -5740,17 +5730,6 @@
     /* call the WH_CBT hook */
 
     release_win_ptr( win );
-
-    if (class && IS_INTRESOURCE(class)) cs.lpszClass = class;
-    else
-    {
-        UNICODE_STRING base_name;
-        base_name.Buffer = base_nameW;
-        base_name.MaximumLength = sizeof(base_nameW);
-        NtUserGetClassName( hwnd, FALSE, &base_name );
-        cs.lpszClass = base_name.Buffer;
-    }
-
     cbtc.hwndInsertAfter = HWND_TOP;
     cbtc.lpcs = &cs;
     if (call_hooks( WH_CBT, HCBT_CREATEWND, (WPARAM)hwnd, (LPARAM)&cbtc, sizeof(cbtc) ))
@@ -5780,12 +5759,15 @@
 
     if (!(win->dwStyle & (WS_CHILD | WS_POPUP))) win->flags |= WIN_NEED_SIZE;
 
-    SERVER_START_REQ( init_window_info )
+    SERVER_START_REQ( set_window_info )
     {
         req->handle     = wine_server_user_handle( hwnd );
+        req->flags      = SET_WIN_STYLE | SET_WIN_EXSTYLE | SET_WIN_INSTANCE | SET_WIN_UNICODE;
         req->style      = win->dwStyle;
         req->ex_style   = win->dwExStyle;
+        req->instance   = wine_server_client_ptr( win->hInstance );
         req->is_unicode = (win->flags & WIN_ISUNICODE) != 0;
+        req->extra_offset = -1;
         wine_server_call( req );
     }
     SERVER_END_REQ;
@@ -5803,9 +5785,9 @@
     }
     else NtUserSetWindowLongPtr( hwnd, GWLP_ID, (ULONG_PTR)cs.hMenu, FALSE );
 
+    win_dpi = NTUSER_DPI_CONTEXT_GET_DPI( win->dpi_context );
     release_win_ptr( win );
 
-    win_dpi = get_dpi_for_window( hwnd );
     if (parent) map_dpi_create_struct( &cs, win_dpi );
 
     context = set_thread_dpi_awareness_context( get_window_dpi_awareness_context( hwnd ));
@@ -5817,8 +5799,18 @@
     if ((cs.style & WS_THICKFRAME) || !(cs.style & (WS_POPUP | WS_CHILD)))
     {
         MINMAXINFO info = get_min_max_info( hwnd );
-        cx = max( min( cx, info.ptMaxTrackSize.x ), info.ptMinTrackSize.x );
-        cy = max( min( cy, info.ptMaxTrackSize.y ), info.ptMinTrackSize.y );
+
+        /* HACK: This code changes the window's size to fit the display. However,
+         * some games (Bayonetta, Dragon's Dogma) will then have the incorrect
+         * render size. So just let windows be too big to fit the display. */
+        if (disable_gamescope_max_size_hack() || !user_driver->pHasWindowManager( "steamcompmgr" ))
+        {
+            cx = min( cx, info.ptMaxTrackSize.x );
+            cy = min( cy, info.ptMaxTrackSize.y );
+        }
+
+        cx = max( cx, info.ptMinTrackSize.x );
+        cy = max( cy, info.ptMinTrackSize.y );
     }
 
     if (cx < 0) cx = 0;
@@ -5896,7 +5888,7 @@
 
     /* Show the window, maximizing or minimizing if needed */
 
-    style = set_window_style_bits( hwnd, 0, WS_MAXIMIZE | WS_MINIMIZE );
+    style = set_window_style( hwnd, 0, WS_MAXIMIZE | WS_MINIMIZE );
     if (style & (WS_MINIMIZE | WS_MAXIMIZE))
     {
         RECT new_pos;
@@ -5947,7 +5939,7 @@
     set_thread_dpi_awareness_context( context );
 
     toplevel = NtUserGetAncestor( hwnd, GA_ROOT );
-    if (toplevel && toplevel != hwnd) update_window_state( toplevel );
+    if (toplevel != hwnd) update_window_state( toplevel );
 
     return hwnd;
 
@@ -5998,19 +5990,6 @@
     return NtUserSetWindowPos( hwnd, 0, rect.left, rect.top, rect.right - rect.left, rect.bottom - rect.top, flags );
 }
 
-BOOL get_present_rect( HWND hwnd, RECT *rect, UINT dpi )
-{
-    UINT dpi_from = get_dpi_for_window( hwnd );
-    WND *win;
-
-    if (!(win = get_win_ptr( hwnd )) || win == WND_OTHER_PROCESS || win == WND_DESKTOP) return FALSE;
-    if (dpi != -1) *rect = map_dpi_rect( win->present_rect, dpi_from, dpi );
-    else *rect = map_rect_virt_to_raw( win->present_rect, dpi_from );
-    release_win_ptr( win );
-
-    return !IsRectEmpty( rect );
-}
-
 /*****************************************************************************
  *           NtUserCallHwnd (win32u.@)
  */
@@ -6022,8 +6001,11 @@
         activate_other_window( hwnd );
         return 0;
 
-    case NtUserCallHwnd_SetForegroundWindowInternal:
-        return set_foreground_window( hwnd, FALSE, TRUE );
+    case NtUserCallHwnd_ArrangeIconicWindows:
+        return arrange_iconic_windows( hwnd );
+
+    case NtUserCallHwnd_DrawMenuBar:
+        return draw_menu_bar( hwnd );
 
     case NtUserCallHwnd_GetDpiForWindow:
         return get_dpi_for_window( hwnd );
@@ -6041,6 +6023,9 @@
         if (!(win_get_flags( hwnd ) & WIN_ISMDICLIENT)) return 0;
         return get_window_long_ptr( hwnd, sizeof(void *), FALSE );
 
+    case NtUserCallHwnd_GetWindowContextHelpId:
+        return get_window_context_help_id( hwnd );
+
     case NtUserCallHwnd_GetWindowDpiAwarenessContext:
         return get_window_dpi_awareness_context( hwnd );
 
@@ -6065,6 +6050,15 @@
     case NtUserCallHwnd_IsWindowVisible:
         return is_window_visible( hwnd );
 
+    case NtUserCallHwnd_SetForegroundWindow:
+        return set_foreground_window( hwnd, FALSE );
+
+    case NtUserCallHwnd_SetProgmanWindow:
+        return HandleToUlong( set_progman_window( hwnd ));
+
+    case NtUserCallHwnd_SetTaskmanWindow:
+        return HandleToUlong( set_taskman_window( hwnd ));
+
     /* temporary exports */
     case NtUserGetFullWindowHandle:
         return HandleToUlong( get_full_window_handle( hwnd ));
@@ -6076,7 +6070,7 @@
         return HandleToUlong( is_current_thread_window( hwnd ));
 
     default:
-        FIXME( "invalid code %u\n", code );
+        FIXME( "invalid code %u\n", (int)code );
         return 0;
     }
 }
@@ -6091,6 +6085,9 @@
     case NtUserCallHwndParam_ClientToScreen:
         return client_to_screen( hwnd, (POINT *)param );
 
+    case NtUserCallHwndParam_EnableWindow:
+        return enable_window( hwnd, param );
+
     case NtUserCallHwndParam_GetChildRect:
         return get_window_rect_rel( hwnd, COORDS_PARENT, (RECT *)param, get_thread_dpi() );
 
@@ -6130,20 +6127,11 @@
     case NtUserCallHwndParam_GetWindowLongPtrW:
         return get_window_long_ptr( hwnd, param, FALSE );
 
-    case NtUserCallHwndParam_GetWindowRect:
-    {
-        struct get_window_rects_params *params = (void *)param;
-        return get_window_rect( hwnd, params->rect, params->dpi );
-    }
-    case NtUserCallHwndParam_GetClientRect:
-    {
-        struct get_window_rects_params *params = (void *)param;
-        return get_client_rect( hwnd, params->rect, params->dpi );
-    }
-    case NtUserCallHwndParam_GetPresentRect:
+    case NtUserCallHwndParam_GetWindowRects:
     {
         struct get_window_rects_params *params = (void *)param;
-        return get_present_rect( hwnd, params->rect, params->dpi );
+        return params->client ? get_client_rect( hwnd, params->rect, params->dpi )
+                              : get_window_rect( hwnd, params->rect, params->dpi );
     }
 
     case NtUserCallHwndParam_GetWindowRelative:
@@ -6158,6 +6146,9 @@
     case NtUserCallHwndParam_IsChild:
         return is_child( hwnd, UlongToHandle(param) );
 
+    case NtUserCallHwndParam_KillSystemTimer:
+        return kill_system_timer( hwnd, param );
+
     case NtUserCallHwndParam_MapWindowPoints:
         {
             struct map_window_points_params *params = (void *)param;
@@ -6180,6 +6171,12 @@
         NtUserSetWindowLongPtr( hwnd, sizeof(void *), param, FALSE );
         return win_set_flags( hwnd, WIN_ISMDICLIENT, 0 );
 
+    case NtUserCallHwndParam_SetWindowContextHelpId:
+        return set_window_context_help_id( hwnd, param );
+
+    case NtUserCallHwndParam_ShowOwnedPopups:
+        return show_owned_popups( hwnd, param );
+
     case NtUserCallHwndParam_SendHardwareInput:
     {
         struct send_hardware_input_params *params = (void *)param;
@@ -6194,8 +6191,8 @@
 
     case NtUserCallHwndParam_GetWinMonitorDpi:
     {
-        UINT raw_dpi, dpi = get_win_monitor_dpi( hwnd, &raw_dpi );
-        return param == MDT_EFFECTIVE_DPI ? dpi : raw_dpi;
+        UINT raw_dpi_num, raw_dpi_den, dpi = get_win_monitor_dpi( hwnd, &raw_dpi_num, &raw_dpi_den );
+        return param == MDT_EFFECTIVE_DPI ? dpi : (raw_dpi_num + raw_dpi_den / 2) / raw_dpi_den;
     }
 
     case NtUserCallHwndParam_SetRawWindowPos:
@@ -6204,8 +6201,15 @@
         return set_raw_window_pos( hwnd, params->rect, params->flags, params->internal );
     }
 
+    /* temporary exports */
+    case NtUserSetWindowStyle:
+        {
+            STYLESTRUCT *style = (void *)param;
+            return set_window_style( hwnd, style->styleNew, style->styleOld );
+        }
+
     default:
-        FIXME( "invalid code %u\n", code );
+        FIXME( "invalid code %u\n", (int)code );
         return 0;
     }
 }
@@ -6296,7 +6300,7 @@
     {
     case WindowProcess:
     case WindowProcess2:
-        if (!get_window_thread( hwnd, &pid )) return NULL;
+        get_window_thread( hwnd, &pid );
         return UlongToHandle( pid );
 
     case WindowThread:
@@ -6389,10 +6393,7 @@
     return ret;
 }
 
-/***********************************************************************
- *            NtUserSetProgmanWindow (win32u.@)
- */
-HWND WINAPI NtUserSetProgmanWindow( HWND hwnd )
+HWND set_progman_window( HWND hwnd )
 {
     SERVER_START_REQ(set_desktop_shell_windows)
     {
@@ -6418,10 +6419,7 @@
     return ret;
 }
 
-/***********************************************************************
- *            NtUserSetTaskmanWindow (win32u.@)
- */
-HWND WINAPI NtUserSetTaskmanWindow( HWND hwnd )
+HWND set_taskman_window( HWND hwnd )
 {
     /* hwnd = MSTaskSwWClass
      *        |-> SysTabControl32
@@ -6436,39 +6434,19 @@
     return hwnd;
 }
 
-/***********************************************************************
- *            NtUserIsChildWindowDpiMessageEnabled (win32u.@)
- */
-BOOL WINAPI NtUserIsChildWindowDpiMessageEnabled( HWND hwnd )
-{
-    FIXME( "%p: stub\n", hwnd );
-    return FALSE;
-}
-
 /*****************************************************************
  *           NtUserGetWindowDisplayAffinity (win32u.@)
  */
 BOOL WINAPI NtUserGetWindowDisplayAffinity( HWND hwnd, DWORD *affinity )
 {
-    FIXME( "%p, %p: stub\n", hwnd, affinity );
+    FIXME("(%p, %p): stub\n", hwnd, affinity);
 
     if (!hwnd || !affinity)
     {
-        RtlSetLastWin32Error( hwnd ? ERROR_NOACCESS : ERROR_INVALID_WINDOW_HANDLE );
+        RtlSetLastWin32Error(hwnd ? ERROR_NOACCESS : ERROR_INVALID_WINDOW_HANDLE);
         return FALSE;
     }
 
     *affinity = WDA_NONE;
     return TRUE;
 }
-
-/*****************************************************************
- *           NtUserModifyUserStartupInfoFlags (win32u.@)
- */
-BOOL WINAPI NtUserModifyUserStartupInfoFlags( DWORD mask, DWORD flags )
-{
-    TRACE( "%#x, %#x.\n", mask, flags );
-
-    set_startup_info_flags( mask, flags );
-    return TRUE;
-}
