diff --git a/dlls/ntdll/signal_arm64.c b/dlls/ntdll/signal_arm64.c
--- a/dlls/ntdll/signal_arm64.c
+++ b/dlls/ntdll/signal_arm64.c
@@ -671,6 +671,14 @@ BOOLEAN WINAPI RtlIsProcessorFeaturePresent( UINT feature )
     return !!(cpu_features_bitmap[feature / 64] & (1ull << (feature % 64)));
 }
 
+static LONG apc_worker_started = 0;
+
+static void apc_worker_thread( void * )
+{
+    ULONG count = 0;
+    while (1) NtSuspendThread( NtCurrentThread(), &count );
+}
+
 static void suspend_remote_breakin( HANDLE thread )
 {
     ULONG count = 0;
@@ -921,6 +929,17 @@ __ASM_GLOBAL_FUNC( RtlUserThreadStart,
  */
 void WINAPI LdrInitializeThunk( CONTEXT *context, ULONG_PTR unk2, ULONG_PTR unk3, ULONG_PTR unk4 )
 {
+    if (NtCurrentTeb()->WowTebOffset && InterlockedCompareExchange( &apc_worker_started, 1, 0 ) == 0)
+    {
+        HANDLE handle;
+        NTSTATUS status = NtCreateThreadEx( &handle, SYNCHRONIZE | THREAD_QUERY_INFORMATION, NULL,
+                                            NtCurrentProcess(), apc_worker_thread, NULL,
+                                            THREAD_CREATE_FLAGS_SKIP_THREAD_ATTACH | THREAD_CREATE_FLAGS_SKIP_LOADER_INIT |
+                                            THREAD_CREATE_FLAGS_HIDE_FROM_DEBUGGER, 0, 0, 0, NULL );
+        if (!status) NtClose( handle );
+        else InterlockedExchange( &apc_worker_started, 0 );
+    }
+
     loader_init( context, (void **)&context->X0 );
     TRACE_(relay)( "\1Starting thread proc %p (arg=%p)\n", (void *)context->X0, (void *)context->X1 );
     NtContinue( context, TRUE );
