diff --git a/dlls/ntdll/signal_x86_64.c b/dlls/ntdll/signal_x86_64.c
--- a/dlls/ntdll/signal_x86_64.c
+++ b/dlls/ntdll/signal_x86_64.c
@@ -866,7 +866,8 @@ __ASM_GLOBAL_FUNC( RtlRaiseException,
                    "movq %rax,0x58(%rdx)\n\t"   /* Context->Dr2 */
                    "movq %rax,0x60(%rdx)\n\t"   /* Context->Dr3 */
                    "movl $1,%r8d\n\t"
-                   "movq %gs:0x60,%rax\n\t"     /* Peb */
+                   "movq %gs:(0x30),%rax\n\t"   /* Teb */
+                   "movq 0x60(%rax),%rax\n\t"   /* Peb */
                    "cmpb $0,0x02(%rax)\n\t"     /* BeingDebugged */
                    "jne 1f\n\t"
                    "call " __ASM_NAME("dispatch_exception") "\n"
