diff --git a/dlls/ntdll/unix/loader.c b/dlls/ntdll/unix/loader.c
--- a/dlls/ntdll/unix/loader.c
+++ b/dlls/ntdll/unix/loader.c
@@ -367,9 +367,13 @@ static WORD get_alt_machine( WORD machine )
 static void set_dll_path(void)
 {
     char *p, *path = getenv( "WINEDLLPATH" );
+    char *be_runtime = getenv( "PROTON_BATTLEYE_RUNTIME" );
+    char *eac_runtime = getenv( "PROTON_EAC_RUNTIME" );
     int i, count = 0;
 
     if (path) for (p = path, count = 1; *p; p++) if (*p == ':') count++;
+    if (be_runtime) count += 2;
+    if (eac_runtime) count += 2;
 
     dll_paths = malloc( (count + 2) * sizeof(*dll_paths) );
     count = 0;
@@ -383,6 +387,38 @@ static void set_dll_path(void)
         free( path );
     }
 
+    if (be_runtime)
+    {
+        const char lib32[] = "/v1/lib/wine/";
+        const char lib64[] = "/v1/lib64/wine/";
+
+        p = malloc( strlen( be_runtime ) + strlen( lib32 ) + 1 );
+        strcpy( p, be_runtime );
+        strcat( p, lib32 );
+        dll_paths[count++] = p;
+
+        p = malloc( strlen( be_runtime ) + strlen( lib64 ) + 1 );
+        strcpy( p, be_runtime );
+        strcat( p, lib64 );
+        dll_paths[count++] = p;
+    }
+
+    if (eac_runtime)
+    {
+        const char lib32[] = "/v2/lib32/";
+        const char lib64[] = "/v2/lib64/";
+
+        p = malloc( strlen( eac_runtime ) + strlen( lib32 ) + 1 );
+        strcpy( p, eac_runtime );
+        strcat( p, lib32 );
+        dll_paths[count++] = p;
+
+        p = malloc( strlen( eac_runtime ) + strlen( lib64 ) + 1 );
+        strcpy( p, eac_runtime );
+        strcat( p, lib64 );
+        dll_paths[count++] = p;
+    }
+
     for (i = 0; i < count; i++) dll_path_maxlen = max( dll_path_maxlen, strlen(dll_paths[i]) );
     dll_paths[count] = NULL;
 }
