diff --git a/dlls/ntdll/unix/loader.c b/dlls/ntdll/unix/loader.c
--- a/dlls/ntdll/unix/loader.c
+++ b/dlls/ntdll/unix/loader.c
@@ -2281,7 +2281,21 @@ DECLSPEC_EXPORT void __wine_main( int argc, char *argv[] )
     main_argv = argv;
 
     init_paths();
-    if (!getenv( "WINELOADERNOEXEC" ) || argc <= 1) check_command_line( argc, argv );
+    if (!getenv( "WINELOADERNOEXEC" ))
+    {
+        check_command_line( argc, argv );
+        if (pre_exec())
+        {
+            static char noexec[] = "WINELOADERNOEXEC=1";
+            char **new_argv = malloc( (argc + 2) * sizeof(*argv) );
+
+            memcpy( new_argv + 1, argv, (argc + 1) * sizeof(*argv) );
+            putenv( noexec );
+            loader_exec( new_argv, current_machine );
+            fatal_error( "could not exec the wine loader\n" );
+        }
+    }
+    else if (argc <= 1) check_command_line( argc, argv );
     unsetenv( "WINELOADERNOEXEC" );
 
 #ifdef RLIMIT_NOFILE
