diff --git a/aclocal.m4 b/aclocal.m4
index 64c8d1e4896a529a4a1536165beb99241592a709..2ba926b490cf1c1e276608bc4087e952dc49e454 100644
--- a/aclocal.m4
+++ b/aclocal.m4
@@ -113,10 +113,10 @@ if test -s conftest.err; then
      cat conftest.err >&AS_MESSAGE_LOG_FD
 fi
 rm -f conftest.err
-AS_IF([test -n "$ac_cflags$ac_libs"],[ac_save_CPPFLAGS=$CPPFLAGS
+ac_save_CPPFLAGS=$CPPFLAGS
 CPPFLAGS="$CPPFLAGS $ac_cflags"
 $6
-CPPFLAGS=$ac_save_CPPFLAGS])
+CPPFLAGS=$ac_save_CPPFLAGS
 AS_VAR_POPDEF([ac_libs])dnl
 AS_VAR_POPDEF([ac_cflags])])dnl
 
@@ -135,10 +135,10 @@ AS_VAR_IF([ac_libs],[],
       [AS_VAR_SET_IF([MINGW_PKG_CONFIG],
       [ac_libs=`$MINGW_PKG_CONFIG --libs [$2] 2>/dev/null`])])
 m4_ifval([$3],[ac_libs=[$]{ac_libs:-"$3"}])
-AS_IF([test -n "$ac_cflags$ac_libs"],[ac_save_CPPFLAGS=$CPPFLAGS
+ac_save_CPPFLAGS=$CPPFLAGS
 CPPFLAGS="$CPPFLAGS $ac_cflags"
 $4
-CPPFLAGS=$ac_save_CPPFLAGS])
+CPPFLAGS=$ac_save_CPPFLAGS
 AS_VAR_POPDEF([ac_libs])dnl
 AS_VAR_POPDEF([ac_cflags])])dnl
 
@@ -149,16 +149,14 @@ dnl
 AC_DEFUN([WINE_EXTLIB_FLAGS],
 [AS_VAR_PUSHDEF([ac_cflags],[[$1]_PE_CFLAGS])dnl
 AS_VAR_PUSHDEF([ac_libs],[[$1]_PE_LIBS])dnl
-AS_VAR_PUSHDEF([ac_enable],[enable_[$2]])dnl
 AC_ARG_VAR(ac_cflags, [C compiler flags for the PE $2, overriding the bundled version])dnl
 AC_ARG_VAR(ac_libs, [Linker flags for the PE $2, overriding the bundled version])dnl
 AS_VAR_IF([ac_libs],[],
   [ac_libs=$3
-   AS_VAR_IF([ac_cflags],[],[ac_cflags=$4],[ac_enable=no])],
-  [ac_enable=no])
+   AS_VAR_IF([ac_cflags],[],[ac_cflags=$4],[enable_$2=no])],
+  [enable_$2=no])
 AS_ECHO(["$as_me:${as_lineno-$LINENO}: $2 cflags: $ac_cflags"]) >&AS_MESSAGE_LOG_FD
 AS_ECHO(["$as_me:${as_lineno-$LINENO}: $2 libs: $ac_libs"]) >&AS_MESSAGE_LOG_FD
-AS_VAR_POPDEF([ac_enable])dnl
 AS_VAR_POPDEF([ac_libs])dnl
 AS_VAR_POPDEF([ac_cflags])])dnl
 
@@ -277,8 +275,7 @@ dnl
 dnl Usage: AC_REQUIRE([WINE_CONFIG_HELPERS])
 dnl
 AC_DEFUN([WINE_CONFIG_HELPERS],
-[AS_VAR_SET([wine_rules],["all:
-	@echo 'Wine build complete.'"])
+[AS_VAR_SET([wine_rules],["all:"])
 AC_SUBST(SUBDIRS,"")
 AC_SUBST(DISABLED_SUBDIRS,"")
 AC_SUBST(CONFIGURE_TARGETS,"")
@@ -291,7 +288,7 @@ wine_fn_config_makefile ()
       no) AS_VAR_APPEND([DISABLED_SUBDIRS],[" $[1]"]) ;;
       yes) ;;
       *aarch64*|*arm*|*i386*|*x86_64*)
-        if test -n "$PE_ARCHS" -a "$PE_ARCHS" != none
+        if test -n "$PE_ARCHS"
         then
             for i in $PE_ARCHS
             do
@@ -305,6 +302,15 @@ wine_fn_config_makefile ()
           programs/*,*\ arm64ec\ *) AS_VAR_APPEND([arm64ec_DISABLED_SUBDIRS],[" $[1]"]) ;;
         esac ;;
     esac
+}
+
+wine_fn_config_symlink ()
+{
+    ac_links=$[@]
+    AS_VAR_APPEND([wine_rules],["
+$ac_links:
+	@./config.status \$[@]"])
+    for f in $ac_links; do AS_VAR_APPEND([CONFIGURE_TARGETS],[" $f"]); done
 }])
 
 dnl **** Define helper function to append a rule to a makefile command list ****
@@ -314,15 +320,26 @@ dnl
 AC_DEFUN([WINE_APPEND_RULE],[AC_REQUIRE([WINE_CONFIG_HELPERS])AS_VAR_APPEND([wine_rules],["
 $1"])])
 
+dnl **** Create symlinks from config.status ****
+dnl
+dnl Usage: WINE_CONFIG_SYMLINK(target,src,enable)
+dnl
+AC_DEFUN([WINE_CONFIG_SYMLINK],[AC_REQUIRE([WINE_CONFIG_HELPERS])dnl
+m4_ifval([$3],[if test $3; then
+])AC_CONFIG_LINKS([$1:$2])dnl
+wine_fn_config_symlink[ $1]m4_ifval([$3],[
+fi])[]dnl
+])])
+
 dnl **** Create a makefile from config.status ****
 dnl
-dnl Usage: WINE_CONFIG_MAKEFILE(file)
+dnl Usage: WINE_CONFIG_MAKEFILE(file,enable,condition)
 dnl
 AC_DEFUN([WINE_CONFIG_MAKEFILE],[AC_REQUIRE([WINE_CONFIG_HELPERS])dnl
-AS_VAR_PUSHDEF([ac_enable],[enable_]m4_bpatsubst(m4_bpatsubst([$1],[.*/\([^/]*\)$],[\1]),[.*\.\(.*16\|vxd\)$],[win16]))dnl
+AS_VAR_PUSHDEF([ac_enable],m4_default([$2],[enable_]m4_bpatsubst([$1],[.*/\([^/]*\)$],[\1])))dnl
 m4_append_uniq([_AC_USER_OPTS],ac_enable,[
 ])dnl
-m4_if(m4_bregexp([$1],[^tools]),[-1],[],[test "x$enable_tools" = xno || ])wine_fn_config_makefile [$1] ac_enable[]dnl
+m4_ifval([$3],[$3 || ])wine_fn_config_makefile [$1] ac_enable[]dnl
 AS_VAR_POPDEF([ac_enable])])
 
 dnl **** Append a file to the .gitignore list ****
