diff --git a/dlls/d3dx9_36/effect.c b/dlls/d3dx9_36/effect.c
index a89dc679a7..06b349673e 100644
--- a/dlls/d3dx9_36/effect.c
+++ b/dlls/d3dx9_36/effect.c
@@ -4432,10 +4432,158 @@ static HRESULT WINAPI d3dx_effect_CloneEffect(ID3DXEffect *iface, IDirect3DDevic
 static HRESULT WINAPI d3dx_effect_SetRawValue(ID3DXEffect *iface, D3DXHANDLE parameter, const void *data,
         UINT byte_offset, UINT bytes)
 {
+<<<<<<< ours
     FIXME("iface %p, parameter %p, data %p, byte_offset %u, bytes %u stub!\n",
             iface, parameter, data, byte_offset, bytes);
 
     return E_NOTIMPL;
+=======
+    struct d3dx_effect *effect = impl_from_ID3DXEffect(iface);
+    struct d3dx_parameter *param = get_valid_parameter(effect, parameter);
+
+    TRACE("iface %p, parameter %p, data %p, byte_offset %u, bytes %u.\n", iface, parameter, data, byte_offset, bytes);
+
+    if (!param)
+    {
+        WARN("Invalid parameter %p specified.\n", parameter);
+        return D3DERR_INVALIDCALL;
+    }
+
+    switch (param->class)
+    {
+        case D3DXPC_VECTOR:
+        {
+            uint8_t *dst_data;
+
+            if (param->columns == 4)
+            {
+                if ((byte_offset + bytes) > param->bytes)
+                {
+                    FIXME("Writing adjacent parameters is currently unsupported.\n");
+                    return E_NOTIMPL;
+                }
+
+                dst_data = param_get_data_and_dirtify(effect, param, !byte_offset ? bytes : param->bytes, TRUE);
+                memcpy(dst_data + byte_offset, data, bytes);
+            }
+            else
+            {
+                unsigned int src_elems = (bytes + 0xf) / sizeof(D3DXVECTOR4);
+                unsigned int dst_stride = sizeof(float) * param->columns;
+                unsigned int dst_elems = max(1, param->element_count);
+                const D3DXVECTOR4 *src_data = data;
+                unsigned int i, bytes_left;
+
+                if (byte_offset)
+                {
+                    FIXME("Setting Vec%u parameters with an offset is currently unsupported.\n", param->columns);
+                    return E_NOTIMPL;
+                }
+
+                if (src_elems > dst_elems)
+                {
+                    FIXME("Writing adjacent parameters is currently unsupported.\n");
+                    return E_NOTIMPL;
+                }
+
+                bytes_left = bytes;
+                dst_data = param_get_data_and_dirtify(effect, param, dst_elems * dst_stride, TRUE);
+                for (i = 0; i < src_elems; ++i)
+                {
+                    memcpy(dst_data + (i * dst_stride), &src_data[i], min(bytes_left, dst_stride));
+                    bytes_left -= sizeof(*src_data);
+                }
+            }
+
+            break;
+        }
+
+        case D3DXPC_MATRIX_ROWS:
+        {
+            D3DXMATRIX *dst_elem;
+            uint8_t *dst_data;
+            unsigned int i;
+
+            if (param->columns != 4 || param->rows != 4)
+            {
+                FIXME("%ux%u matrix parameters are currently unsupported.\n", param->rows, param->columns);
+                return E_NOTIMPL;
+            }
+
+            if (byte_offset & 0x3f || bytes & 0x3f)
+            {
+                FIXME("Partial matrix updates are currently unsupported.\n");
+                return E_NOTIMPL;
+            }
+
+            if ((byte_offset + bytes) > param->bytes)
+            {
+                FIXME("Writing adjacent parameters is currently unsupported.\n");
+                return E_NOTIMPL;
+            }
+
+            dst_data = param_get_data_and_dirtify(effect, param, !byte_offset ? bytes : param->bytes, TRUE);
+            dst_elem = (D3DXMATRIX *)(dst_data + byte_offset);
+            for (i = 0; i < (bytes / sizeof(D3DXMATRIX)); ++i)
+                D3DXMatrixTranspose(&dst_elem[i], &((const D3DXMATRIX *)data)[i]);
+            break;
+        }
+
+        case D3DXPC_STRUCT:
+        {
+            const uint8_t *cur_param_data = data;
+            UINT bytes_left = bytes;
+            unsigned int i;
+            HRESULT hr;
+
+            if (byte_offset)
+            {
+                FIXME("Setting structure members at an offset is currently unsupported.\n");
+                return E_NOTIMPL;
+            }
+
+            for (i = 0; i < param->member_count; ++i)
+            {
+                UINT member_size = param->members[i].class == D3DXPC_MATRIX_ROWS ? 64 : 16;
+                unsigned int member_elems = max(1, param->members[i].element_count);
+                struct d3dx_parameter *member = &param->members[i];
+
+                if (member->class != D3DXPC_MATRIX_ROWS && member->class != D3DXPC_VECTOR)
+                {
+                    FIXME("Unhandled structure member parameter class %s.\n", debug_d3dxparameter_class(member->class));
+                    return E_NOTIMPL;
+                }
+
+                member_size *= member_elems;
+                hr = iface->lpVtbl->SetRawValue(iface, (D3DXHANDLE)member, (const void *)cur_param_data, 0,
+                        min(member_size, bytes_left));
+                if (FAILED(hr))
+                {
+                    WARN("Failed with hr %#lx.\n", hr);
+                    return hr;
+                }
+
+                if (bytes_left <= member_size)
+                {
+                    bytes_left = 0;
+                    break;
+                }
+                cur_param_data += member_size;
+                bytes_left -= member_size;
+            }
+
+            if (bytes_left)
+                FIXME("%u bytes were leftover, might have attempted to write an adjacent parameter.\n", bytes_left);
+            break;
+        }
+
+        default:
+            FIXME("Unhandled parameter class %s.\n", debug_d3dxparameter_class(param->class));
+            return E_NOTIMPL;
+    }
+
+    return D3D_OK;
+>>>>>>> theirs
 }
 #endif
 
diff --git a/dlls/d3dx9_36/tests/effect.c b/dlls/d3dx9_36/tests/effect.c
index c55019f0f7..5f8bb4539b 100644
--- a/dlls/d3dx9_36/tests/effect.c
+++ b/dlls/d3dx9_36/tests/effect.c
@@ -8890,6 +8890,299 @@ static void test_effect_parameter_block(void)
     DestroyWindow(window);
 }
 
+<<<<<<< ours
+=======
+#if D3DX_SDK_VERSION >= 27
+static void test_effect_set_raw_value(IDirect3DDevice9 *device)
+{
+    static const char test_set_raw_value_shader[] =
+        "bool b;\n"
+        "bool b_2[2];\n"
+        "bool2 b2;\n"
+        "bool2 b2_2[2];\n"
+        "bool4 b4;\n"
+        "bool4 b4_2[2];\n"
+        "bool2x2 b22;\n"
+        "bool2x2 b22_2[2];\n"
+        "bool4x4 b44;\n"
+        "bool4x4 b44_2[2];\n"
+        "int i;\n"
+        "int i_2[2];\n"
+        "int2 i2;\n"
+        "int2 i2_2[2];\n"
+        "int4 i4;\n"
+        "int4 i4_2[2];\n"
+        "int2x2 i22;\n"
+        "int2x2 i22_2[2];\n"
+        "int4x4 i44;\n"
+        "int4x4 i44_2[2];\n"
+        "float f;\n"
+        "float f_2[2];\n"
+        "float2 f2;\n"
+        "float2 f2_2[2];\n"
+        "float4 f4;\n"
+        "float4 f4_2[2];\n"
+        "float2x2 f22;\n"
+        "float2x2 f22_2[2];\n"
+        "float4x4 f44;\n"
+        "float4x4 f44_2[2];\n"
+        "technique t { pass p { } };\n";
+    static const char *param_type_str[] = { "b", "i", "f" };
+    static const struct
+    {
+        const char *suffix;
+        UINT value_offset;
+        UINT value_bytes;
+        union
+        {
+            float f[32];
+            DWORD dword[32];
+        } value;
+        union
+        {
+            float f[32];
+            DWORD dword[32];
+        } expected_value;
+        BOOL todo_hr;
+    } tests[] =
+    {
+        { NULL, 0, 4,
+          { .f = { 1.0f } },
+          { .f = { 1.0f } },
+          .todo_hr = TRUE
+        },
+        { "_2", 0, 8,
+          { .f = { 1.0f, 2.0f } },
+          { .f = { 1.0f, 0.0f } },
+          .todo_hr = TRUE
+        },
+        { "_2", 0, 20,
+          { .f = { 1.0f, 2.0f, 0.0f, 0.0f, 3.0f } },
+          { .f = { 1.0f, 3.0f, 0.0f, 0.0f, 0.0f } },
+          .todo_hr = TRUE
+        },
+        /* Offset of 4, nothing gets set. */
+        { "_2", 4, 4,
+          { .f = { 1.0f } },
+          { .f = { 0.0f, 0.0f } },
+          .todo_hr = TRUE
+        },
+        /*
+         * To set the second value in the array, we need to start at an offset
+         * of 16.
+         */
+        { "_2", 16, 4,
+          { .f = { 2.0f } },
+          { .f = { 0.0f, 2.0f } },
+          .todo_hr = TRUE
+        },
+        /* 5. */
+        { "2", 0, 8,
+          { .f = { 0.0f, 1.0f } },
+          { .f = { 0.0f, 1.0f } },
+        },
+        { "2_2", 0, 16,
+          { .f = { 0.0f, 1.0f, 0.0f, 2.0f } },
+          { .f = { 0.0f, 1.0f, 0.0f, 0.0f } },
+        },
+        { "2_2", 0, 24,
+          { .f = { 0.0f, 1.0f, 2.0f, 3.0f, 4.0f, 0.0f } },
+          { .f = { 0.0f, 1.0f, 4.0f, 0.0f } },
+        },
+        { "4", 0, 16,
+          { .f = { 1.0f, 2.0f, 0.0f, 3.0f } },
+          { .f = { 1.0f, 2.0f, 0.0f, 3.0f } },
+        },
+        { "4_2", 0, 32,
+          { .f = { 2.0f, 0.0f, 0.0f, 8.0f, 3.0f, 4.0f, 0.0f, 5.0f } },
+          { .f = { 2.0f, 0.0f, 0.0f, 8.0f, 3.0f, 4.0f, 0.0f, 5.0f } },
+        },
+        /* 10. */
+        { "22", 0, 64,
+          { .f = {  1.0f,  2.0f,  3.0f,  4.0f,
+                    5.0f,  6.0f,  7.0f,  0.0f,
+                    8.0f,  9.0f, 10.0f, 11.0f,
+                   12.0f, 13.0f, 14.0f,  0.0f }
+          },
+          { .f = { 1.0f, 5.0f, 2.0f, 6.0f } },
+          .todo_hr = TRUE
+        },
+        { "22_2", 0, 128,
+          { .f = {  1.0f,  2.0f,  3.0f,  4.0f,
+                    5.0f,  6.0f,  7.0f,  0.0f,
+                    8.0f,  9.0f, 10.0f, 11.0f,
+                   12.0f, 13.0f, 14.0f,  0.0f,
+                   /* Matrix 2. */
+                   15.0f, 16.0f, 17.0f, 18.0f,
+                   19.0f, 20.0f, 21.0f, 22.0f,
+                   23.0f, 24.0f, 25.0f, 26.0f,
+                   27.0f, 28.0f, 29.0f, 30.0f }
+          },
+          { .f = { 1.0f, 5.0f, 2.0f, 6.0f, 15.0f, 19.0f, 16.0f, 20.0f } },
+          .todo_hr = TRUE
+        },
+        { "44", 0, 64,
+          { .f = {  1.0f,  2.0f,  3.0f,  4.0f,
+                    5.0f,  6.0f,  7.0f,  0.0f,
+                    8.0f,  9.0f, 10.0f, 11.0f,
+                   12.0f, 13.0f, 14.0f,  0.0f }
+          },
+          { .f = {  1.0f,  5.0f,  8.0f, 12.0f,
+                    2.0f,  6.0f,  9.0f, 13.0f,
+                    3.0f,  7.0f, 10.0f, 14.0f,
+                    4.0f,  0.0f, 11.0f,  0.0f }
+          },
+        },
+        { "44_2", 0, 128,
+          { .f = {  1.0f,  2.0f,  3.0f,  4.0f,
+                    5.0f,  6.0f,  7.0f,  0.0f,
+                    8.0f,  9.0f, 10.0f, 11.0f,
+                   12.0f, 13.0f, 14.0f,  0.0f,
+                   /* Matrix 2. */
+                   15.0f, 16.0f, 17.0f, 18.0f,
+                   19.0f, 20.0f, 21.0f, 22.0f,
+                   23.0f, 24.0f, 25.0f, 26.0f,
+                   27.0f, 28.0f, 29.0f, 30.0f }
+          },
+          { .f = {  1.0f,  5.0f,  8.0f, 12.0f,
+                    2.0f,  6.0f,  9.0f, 13.0f,
+                    3.0f,  7.0f, 10.0f, 14.0f,
+                    4.0f,  0.0f, 11.0f,  0.0f,
+                   /* Matrix 2. */
+                   15.0f, 19.0f, 23.0f, 27.0f,
+                   16.0f, 20.0f, 24.0f, 28.0f,
+                   17.0f, 21.0f, 25.0f, 29.0f,
+                   18.0f, 22.0f, 26.0f, 30.0f }
+          },
+        },
+        /* Set second element. */
+        { "44_2", 64, 64,
+          { .f = {  1.0f,  2.0f,  3.0f,  4.0f,
+                    5.0f,  6.0f,  7.0f,  0.0f,
+                    8.0f,  9.0f, 10.0f, 11.0f,
+                   12.0f, 13.0f, 14.0f,  0.0f }
+          },
+          { .f = {  0.0f,  0.0f,  0.0f,  0.0f,
+                    0.0f,  0.0f,  0.0f,  0.0f,
+                    0.0f,  0.0f,  0.0f,  0.0f,
+                    0.0f,  0.0f,  0.0f,  0.0f,
+                   /* Matrix 2. */
+                    1.0f,  5.0f,  8.0f, 12.0f,
+                    2.0f,  6.0f,  9.0f, 13.0f,
+                    3.0f,  7.0f, 10.0f, 14.0f,
+                    4.0f,  0.0f, 11.0f,  0.0f }
+          },
+        },
+    };
+    static const DWORD test_int_val[] = { 1, 2, 3, 4 };
+    D3DXPARAMETER_DESC param_desc;
+    D3DXHANDLE param, param2;
+    unsigned int i, j, k;
+    char param_name[64];
+    ID3DXEffect *effect;
+    ULONG refcount;
+    DWORD tmp[32];
+    HRESULT hr;
+
+    hr = D3DXCreateEffect(device, test_set_raw_value_shader,
+            sizeof(test_set_raw_value_shader), NULL, NULL, 0, NULL, &effect, NULL);
+    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+    for (i = 0; i < ARRAY_SIZE(tests); ++i)
+    {
+        for (j = 0; j < ARRAY_SIZE(param_type_str); ++j)
+        {
+            strcpy(param_name, param_type_str[j]);
+            if (tests[i].suffix)
+                strcat(param_name, tests[i].suffix);
+
+            winetest_push_context("Test %u (param %s)", i, param_name);
+            param = effect->lpVtbl->GetParameterByName(effect, NULL, param_name);
+
+            hr = effect->lpVtbl->GetParameterDesc(effect, param, &param_desc);
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+            /* Clear our values first. */
+            memset(tmp, 0, sizeof(tmp));
+            hr = effect->lpVtbl->SetValue(effect, param, tmp, sizeof(tmp));
+            ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+            hr = effect->lpVtbl->SetRawValue(effect, param, tests[i].value.dword, tests[i].value_offset,
+                    tests[i].value_bytes);
+            todo_wine_if(tests[i].todo_hr) ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+            if (SUCCEEDED(hr))
+            {
+                unsigned int unexpected_values = 0;
+
+                memset(tmp, 0xff, sizeof(tmp));
+                hr = effect->lpVtbl->GetValue(effect, param, tmp, sizeof(tmp));
+                ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+                for (k = 0; k < (param_desc.Bytes / sizeof(*tmp)); ++k)
+                {
+                    const DWORD *expected_val = &tests[i].expected_value.dword[k];
+                    DWORD *ret_val = &tmp[k];
+
+                    if (param_desc.Type == D3DXPT_BOOL && (!!(*expected_val) != (*ret_val)))
+                        unexpected_values++;
+                    else if (param_desc.Type != D3DXPT_BOOL && memcmp(ret_val, expected_val, sizeof(*ret_val)))
+                        unexpected_values++;
+                }
+
+                for (; k < ARRAY_SIZE(tmp); ++k)
+                {
+                    if (tmp[k] != 0xffffffffu)
+                        unexpected_values++;
+                }
+
+                ok(!unexpected_values, "Got %u unexpected values.\n", unexpected_values);
+            }
+            winetest_pop_context();
+        }
+    }
+
+    /* Test passing a NULL parameter. */
+    hr = effect->lpVtbl->SetRawValue(effect, NULL, tmp, 4, 0);
+    ok(hr == D3DERR_INVALIDCALL, "Unexpected hr %#lx.\n", hr);
+
+    param = effect->lpVtbl->GetParameterByName(effect, NULL, "i4");
+    param2 = effect->lpVtbl->GetParameterByName(effect, NULL, "i44");
+
+    /* Test setting with a size of 0. */
+    hr = effect->lpVtbl->SetRawValue(effect, param, tmp, 0, 0);
+    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+    hr = effect->lpVtbl->SetRawValue(effect, param2, tmp, 0, 0);
+    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+    param = effect->lpVtbl->GetParameterByName(effect, NULL, "i4");
+    param2 = effect->lpVtbl->GetParameterByName(effect, NULL, "i4_2");
+
+    /* Test setting adjacent variables. */
+    memset(tmp, 0, sizeof(tmp));
+    hr = effect->lpVtbl->SetValue(effect, param, tmp, sizeof(tmp));
+    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+    hr = effect->lpVtbl->SetValue(effect, param2, tmp, sizeof(tmp));
+    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+    /* An offset of 16 bytes on variable i4 sets adjacent variable i4_2. */
+    hr = effect->lpVtbl->SetRawValue(effect, param, test_int_val, 16, sizeof(test_int_val));
+    todo_wine ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+
+    hr = effect->lpVtbl->GetValue(effect, param, tmp, sizeof(tmp));
+    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+    ok(!tmp[0], "Unexpected value.\n");
+
+    hr = effect->lpVtbl->GetValue(effect, param2, tmp, sizeof(tmp));
+    ok(hr == D3D_OK, "Unexpected hr %#lx.\n", hr);
+    todo_wine ok(!memcmp(tmp, test_int_val, sizeof(test_int_val)), "Got unexpected values.\n");
+
+    refcount = effect->lpVtbl->Release(effect);
+    ok(!refcount, "Unexpected refcount %lu.\n", refcount);
+}
+#endif /* D3DX_SDK_VERSION >= 27 */
+
+>>>>>>> theirs
 START_TEST(effect)
 {
     IDirect3DDevice9 *device;
