From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: WCP <wcp@example.invalid>
Date: Wed, 4 Mar 2026 00:00:00 +0000
Subject: [PATCH] odbc32: SQLGetData support ODBC v2.0

---
 dlls/odbc32/proxyodbc.c | 23 +++++++++++++++++++++++
 1 file changed, 23 insertions(+)

diff --git a/dlls/odbc32/proxyodbc.c b/dlls/odbc32/proxyodbc.c
index 5310f52..2b22b7a 100644
--- a/dlls/odbc32/proxyodbc.c
+++ b/dlls/odbc32/proxyodbc.c
@@ -323,6 +323,16 @@ static CRITICAL_SECTION_DEBUG loader_cs_debug =
 };
 static CRITICAL_SECTION loader_cs = { &loader_cs_debug, -1, 0, 0, 0, 0 };
 
+static struct object *find_object_type(SQLSMALLINT type, struct object *object)
+{
+    while (object && object->type != type)
+    {
+        object = object->parent;
+    }
+
+    return object;
+}
+
 static struct
 {
     UINT32 count;
@@ -2489,7 +2499,20 @@ static SQLRETURN get_data_win32( struct statement *stmt, SQLUSMALLINT column, SQ
                                  SQLLEN buflen, SQLLEN *retlen )
 {
     if (stmt->hdr.win32_funcs->SQLGetData)
+    {
+        struct connection *conn = (struct connection *)find_object_type(SQL_HANDLE_DBC, stmt->hdr.parent);
+        if (conn && conn->driver_odbc_ver < 0x300)
+        {
+            if (type == SQL_C_TYPE_TIME)
+                type = SQL_C_TIME;
+            else if (type == SQL_C_TYPE_DATE)
+                type = SQL_C_DATE;
+            else if (type == SQL_C_TYPE_TIMESTAMP)
+                type = SQL_C_TIMESTAMP;
+        }
+
         return stmt->hdr.win32_funcs->SQLGetData( stmt->hdr.win32_handle, column, type, value, buflen, retlen );
+    }
     return SQL_ERROR;
 }
 
-- 
2.51.0

