diff --git a/dlls/winmm/mci.c b/dlls/winmm/mci.c
index 959801aa83..0e26ed5366 100644
--- a/dlls/winmm/mci.c
+++ b/dlls/winmm/mci.c
@@ -247,6 +247,14 @@ static int MCI_MapMsgAtoW(UINT msg, DWORD_PTR dwParam1, DWORD_PTR *dwParam2)
             MCI_ANIM_OPEN_PARMSW *mci_openW;
             DWORD_PTR *ptr;
 
+            if ((dwParam1 & (MCI_OPEN_TYPE|MCI_OPEN_TYPE_ID)) == MCI_OPEN_TYPE &&
+                (GetVersion() & 0x80000000) && IsBadStringPtrA(mci_openA->lpstrDeviceType, -1))
+                return MCIERR_MISSING_COMMAND_STRING;
+
+            if ((dwParam1 & (MCI_OPEN_TYPE|MCI_OPEN_TYPE_ID)) == MCI_OPEN_TYPE &&
+                (GetVersion() & 0x80000000) && IsBadStringPtrA(mci_openA->lpstrDeviceType, -1))
+                return MCIERR_MISSING_COMMAND_STRING;
+
             ptr = malloc(sizeof(DWORD_PTR) + sizeof(*mci_openW));
             if (!ptr) return -1;
 
@@ -2286,7 +2294,6 @@ DWORD WINAPI mciSendCommandA(MCIDEVICEID wDevID, UINT wMsg, DWORD_PTR dwParam1,
 
     TRACE("(%08x, %s, %08Ix, %08Ix)\n",
 	  wDevID, MCI_MessageToString(wMsg), dwParam1, dwParam2);
-
     mapped = MCI_MapMsgAtoW(wMsg, dwParam1, &dwParam2);
     if (mapped == -1)
     {
@@ -2294,8 +2301,12 @@ DWORD WINAPI mciSendCommandA(MCIDEVICEID wDevID, UINT wMsg, DWORD_PTR dwParam1,
         return MCIERR_OUT_OF_MEMORY;
     }
     ret = mciSendCommandW(wDevID, wMsg, dwParam1, dwParam2);
+<<<<<<< ours
     if (mapped)
         MCI_UnmapMsgAtoW(wMsg, dwParam1, dwParam2, ret);
+=======
+    MCI_UnmapMsgAtoW(wMsg, dwParam1, dwParam2, ret);
+>>>>>>> theirs
     return ret;
 }
 
