diff --git a/include/aviriff.h b/include/aviriff.h
index 4abac3dc7138c3819025b1d0b4c26924cd465338..9fea63a6f0e96c551d3c608036fd0f6984717afb 100644
--- a/include/aviriff.h
+++ b/include/aviriff.h
@@ -21,7 +21,7 @@
 #define AVIRIFF_H
 
 /* All definitions here are packed structures of on-disk formats */
-#pragma pack(push,2)
+#include <pshpack2.h>
 
 typedef struct _riffchunk
 {
@@ -324,6 +324,6 @@ typedef struct _avifieldindex_chunk {
     } aIndex[ANYSIZE_ARRAY];
 } AVIFIELDINDEX, * PAVIFIELDINDEX;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif  /* AVIRIFF_H */
