diff --git a/include/basetsd.h b/include/basetsd.h
index b3587da03421cb2fa3511377f10503b77adf28ba..3f1f8557aa47aa07db6f2832e9dad102a9bb6482 100644
--- a/include/basetsd.h
+++ b/include/basetsd.h
@@ -95,23 +95,18 @@ typedef unsigned int     ULONG32, *PULONG32;
 typedef unsigned int     DWORD32, *PDWORD32;
 
 #ifdef _MSC_VER
-typedef signed __int64   INT64;
-typedef unsigned __int64 UINT64;
-typedef signed __int64   LONG64;
-typedef unsigned __int64 ULONG64;
-typedef unsigned __int64 DWORD64;
+typedef signed __int64   INT64, *PINT64;
+typedef unsigned __int64 UINT64, *PUINT64;
+typedef signed __int64   LONG64, *PLONG64;
+typedef unsigned __int64 ULONG64, *PULONG64;
+typedef unsigned __int64 DWORD64, *PDWORD64;
 #else
-typedef signed __int64   DECLSPEC_ALIGN(8) INT64;
-typedef unsigned __int64 DECLSPEC_ALIGN(8) UINT64;
-typedef signed __int64   DECLSPEC_ALIGN(8) LONG64;
-typedef unsigned __int64 DECLSPEC_ALIGN(8) ULONG64;
-typedef unsigned __int64 DECLSPEC_ALIGN(8) DWORD64;
+typedef signed __int64   DECLSPEC_ALIGN(8) INT64, *PINT64;
+typedef unsigned __int64 DECLSPEC_ALIGN(8) UINT64, *PUINT64;
+typedef signed __int64   DECLSPEC_ALIGN(8) LONG64, *PLONG64;
+typedef unsigned __int64 DECLSPEC_ALIGN(8) ULONG64, *PULONG64;
+typedef unsigned __int64 DECLSPEC_ALIGN(8) DWORD64, *PDWORD64;
 #endif
-typedef INT64 *PINT64;
-typedef UINT64 *PUINT64;
-typedef LONG64 *PLONG64;
-typedef ULONG64 *PULONG64;
-typedef DWORD64 *PDWORD64;
 
 /* Basic pointer-sized integer types */
 
@@ -165,7 +160,7 @@ typedef unsigned int UHALF_PTR, *PUHALF_PTR;
 
 #if !defined(__midl) && !defined(__WIDL__)
 
-#if !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES) && !defined(WINE_UNIX_LIB)
+#if !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES)
 
 static inline unsigned long HandleToULong(const void *h)
 {
@@ -211,7 +206,7 @@ static inline int PtrToLong(const void *p)
 }
 
 
-#endif /* !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES) && !defined(WINE_UNIX_LIB) */
+#endif /* !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES) */
 
 static inline void *ULongToHandle(ULONG32 ul)
 {
