diff --git a/include/bcrypt.h b/include/bcrypt.h
index 26550bb4b96e8372bbf083b3e9244213f860b205..bb6ed5c16254363f7a7951b5e528a068d452eb00 100644
--- a/include/bcrypt.h
+++ b/include/bcrypt.h
@@ -89,11 +89,9 @@ typedef LONG NTSTATUS;
 #define BCRYPT_DES_ALGORITHM        L"DES"
 #define BCRYPT_DH_ALGORITHM         L"DH"
 #define BCRYPT_DSA_ALGORITHM        L"DSA"
-#define BCRYPT_ECDH_ALGORITHM       L"ECDH"
 #define BCRYPT_ECDH_P256_ALGORITHM  L"ECDH_P256"
 #define BCRYPT_ECDH_P384_ALGORITHM  L"ECDH_P384"
 #define BCRYPT_ECDH_P521_ALGORITHM  L"ECDH_P521"
-#define BCRYPT_ECDSA_ALGORITHM      L"ECDSA"
 #define BCRYPT_ECDSA_P256_ALGORITHM L"ECDSA_P256"
 #define BCRYPT_ECDSA_P384_ALGORITHM L"ECDSA_P384"
 #define BCRYPT_ECDSA_P521_ALGORITHM L"ECDSA_P521"
@@ -118,12 +116,6 @@ typedef LONG NTSTATUS;
 #define BCRYPT_CHAIN_MODE_CCM       L"ChainingModeCCM"
 #define BCRYPT_CHAIN_MODE_GCM       L"ChainingModeGCM"
 
-#define BCRYPT_ECC_CURVE_NAME             L"ECCCurveName"
-#define BCRYPT_ECC_CURVE_BRAINPOOLP256R1  L"brainpoolP256r1"
-#define BCRYPT_ECC_CURVE_SECP256R1        L"secP256r1"
-#define BCRYPT_ECC_CURVE_SECP384R1        L"secP384r1"
-#define BCRYPT_ECC_CURVE_SECP521R1        L"secP521r1"
-
 #define BCRYPT_KDF_HASH             L"HASH"
 #define BCRYPT_KDF_HMAC             L"HMAC"
 #define BCRYPT_KDF_TLS_PRF          L"TLS_PRF"
@@ -182,11 +174,8 @@ static const WCHAR BCRYPT_AES_ALGORITHM[] = {'A','E','S',0};
 static const WCHAR BCRYPT_DES_ALGORITHM[] = {'D','E','S',0};
 static const WCHAR BCRYPT_DH_ALGORITHM[] = {'D','H',0};
 static const WCHAR BCRYPT_DSA_ALGORITHM[] = {'D','S','A',0};
-static const WCHAR BCRYPT_ECDH_ALGORITHM[] = {'E','C','D','H',0};
 static const WCHAR BCRYPT_ECDH_P256_ALGORITHM[] = {'E','C','D','H','_','P','2','5','6',0};
 static const WCHAR BCRYPT_ECDH_P384_ALGORITHM[] = {'E','C','D','H','_','P','3','8','4',0};
-static const WCHAR BCRYPT_ECDH_P521_ALGORITHM[] = {'E','C','D','H','_','P','5','2','1',0};
-static const WCHAR BCRYPT_ECDSA_ALGORITHM[] = {'E','C','D','S','A',0};
 static const WCHAR BCRYPT_ECDSA_P256_ALGORITHM[] = {'E','C','D','S','A','_','P','2','5','6',0};
 static const WCHAR BCRYPT_ECDSA_P384_ALGORITHM[] = {'E','C','D','S','A','_','P','3','8','4',0};
 static const WCHAR BCRYPT_ECDSA_P521_ALGORITHM[] = {'E','C','D','S','A','_','P','5','2','1',0};
@@ -211,12 +200,6 @@ static const WCHAR BCRYPT_CHAIN_MODE_CFB[] = {'C','h','a','i','n','i','n','g','M
 static const WCHAR BCRYPT_CHAIN_MODE_CCM[] = {'C','h','a','i','n','i','n','g','M','o','d','e','C','C','M',0};
 static const WCHAR BCRYPT_CHAIN_MODE_GCM[] = {'C','h','a','i','n','i','n','g','M','o','d','e','G','C','M',0};
 
-static const WCHAR BCRYPT_ECC_CURVE_NAME[] = {'E','C','C','C','u','r','v','e','N','a','m','e',0};
-static const WCHAR BCRYPT_ECC_CURVE_BRAINPOOLP256R1[] = {'b','r','a','i','n','p','o','o','l','P','2','5','6','r','1',0};
-static const WCHAR BCRYPT_ECC_CURVE_SECP256R1[] = {'s','e','c','P','2','5','6','r','1',0};
-static const WCHAR BCRYPT_ECC_CURVE_SECP384R1[] = {'s','e','c','P','3','8','4','r','1',0};
-static const WCHAR BCRYPT_ECC_CURVE_SECP521R1[] = {'s','e','c','P','5','2','1','r','1',0};
-
 static const WCHAR BCRYPT_KDF_HASH[] = {'H','A','S','H',0};
 static const WCHAR BCRYPT_KDF_HMAC[] = {'H','M','A','C',0};
 static const WCHAR BCRYPT_KDF_TLS_PRF[] = {'T','L','S','_','P','R','F',0};
