diff --git a/include/bthdef.h b/include/bthdef.h
index b61d5b033d32b92efec8e10dde66c8dc00631721..d71b2e005e73df6b6b2a2bb1c96b7be3018a08c2 100644
--- a/include/bthdef.h
+++ b/include/bthdef.h
@@ -29,14 +29,7 @@ DEFINE_GUID( GUID_BTHPORT_DEVICE_INTERFACE, 0x850302a, 0xb344, 0x4fda, 0x9b, 0xe
 DEFINE_GUID( GUID_BLUETOOTH_RADIO_INTERFACE, 0x92383b0e, 0xf90e, 0x4ac9, 0x8d, 0x44, 0x8c, 0x2d,
              0x0d, 0x0e, 0xbd, 0xa2 );
 
-DEFINE_GUID( GUID_BLUETOOTH_RADIO_IN_RANGE, 0xea3b5b82, 0x26ee, 0x450e, 0xb0, 0xd8, 0xd2, 0x6f,
-             0xe3, 0x0a, 0x38, 0x69 );
-
-DEFINE_GUID( GUID_BLUETOOTH_RADIO_OUT_OF_RANGE, 0xe28867c9, 0xc2aa, 0x4ced, 0xb9, 0x69, 0x45, 0x70,
-             0x86, 0x60, 0x37, 0xc4);
-
 typedef ULONG BTH_COD;
-typedef ULONGLONG BTH_ADDR, *PBTH_ADDR;
 
 #define BTH_MAX_NAME_SIZE  (248)
 
@@ -59,24 +52,6 @@ typedef struct _BTH_DEVICE_INFO
     CHAR name[BTH_MAX_NAME_SIZE];
 } BTH_DEVICE_INFO, *PBTH_DEVICE_INFO;
 
-/* Buffer for GUID_BLUETOOTH_RADIO_IN_RANGE events. */
-typedef struct _BTH_RADIO_IN_RANGE
-{
-    BTH_DEVICE_INFO deviceInfo;
-    ULONG previousDeviceFlags;
-} BTH_RADIO_IN_RANGE, *PBTH_RADIO_IN_RANGE;
-
-
-typedef enum _AUTHENTICATION_REQUIREMENTS {
-    MITMProtectionNotRequired = 0x00,
-    MITMProtectionRequired = 0x01,
-    MITMProtectionNotRequiredBonding = 0x02,
-    MITMProtectionRequiredBonding = 0x03,
-    MITMProtectionNotRequiredGeneralBonding = 0x04,
-    MITMProtectionRequiredGeneralBonding = 0x05,
-    MITMProtectionNotDefined = 0xff
-} AUTHENTICATION_REQUIREMENTS;
-
 #ifdef __cplusplus
 }
 #endif
