diff --git a/include/combaseapi.h b/include/combaseapi.h
index 7343a72fbbbefdf5da94e71fc8821aeac199ee29..41461f22e98972b1549eadbab853bd6a4b129f0d 100644
--- a/include/combaseapi.h
+++ b/include/combaseapi.h
@@ -53,15 +53,6 @@ enum AgileReferenceOptions
 HRESULT WINAPI CoDecodeProxy(DWORD client_pid, UINT64 proxy_addr, ServerInformation *server_info);
 HRESULT WINAPI RoGetAgileReference(enum AgileReferenceOptions options, REFIID riid, IUnknown *obj, IAgileReference **agile_reference);
 
-#ifdef __cplusplus
-extern "C++" template<typename T> void **IID_PPV_ARGS_Helper(T **obj)
-{
-    (void)static_cast<IUnknown *>(*obj);
-    return reinterpret_cast<void **>(obj);
-}
-#define IID_PPV_ARGS(obj) __uuidof(**(obj)), IID_PPV_ARGS_Helper(obj)
-#endif /* __cplusplus */
-
 #ifdef __cplusplus
 }
 #endif
