diff --git a/include/commctrl.h b/include/commctrl.h
index 5a67c1284a0a2597c2ffa14b54749500a7940b72..c84dce3702c274c5f655759f989cf918ba10b245 100644
--- a/include/commctrl.h
+++ b/include/commctrl.h
@@ -59,7 +59,7 @@ enum _LI_METRIC
 WINCOMMCTRLAPI HRESULT WINAPI LoadIconWithScaleDown(HINSTANCE, const WCHAR *, int, int, HICON *);
 WINCOMMCTRLAPI HRESULT WINAPI LoadIconMetric(HINSTANCE, const WCHAR *, int, HICON *);
 
-#define COMCTL32_VERSION                6  /* dll version */
+#define COMCTL32_VERSION                5  /* dll version */
 
 #define ICC_LISTVIEW_CLASSES   0x00000001  /* listview, header */
 #define ICC_TREEVIEW_CLASSES   0x00000002  /* treeview, tooltips */
@@ -2293,7 +2293,7 @@ static const WCHAR WC_PAGESCROLLERW[] = { 'S','y','s','P','a','g','e','r',0 };
 #define PGN_CALCSIZE            (PGN_FIRST-2)
 #define PGN_HOTITEMCHANGE       (PGN_FIRST-3)
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 typedef struct
 {
@@ -2306,7 +2306,7 @@ typedef struct
     INT  iScroll;
 } NMPGSCROLL, *LPNMPGSCROLL;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef struct
 {
@@ -2830,14 +2830,14 @@ typedef struct tagNMTVASYNCDRAW
     int iRetImageIndex;
 } NMTVASYNCDRAW;
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 typedef struct tagTVKEYDOWN
 {
     NMHDR hdr;
     WORD wVKey;
     UINT flags;
 } NMTVKEYDOWN, *LPNMTVKEYDOWN;
-#pragma pack(pop)
+#include <poppack.h>
 
 #define TV_KEYDOWN      NMTVKEYDOWN
 
@@ -3692,14 +3692,14 @@ typedef struct tagLVDISPINFOW
 #define LV_DISPINFOA	NMLVDISPINFOA
 #define LV_DISPINFOW	NMLVDISPINFOW
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 typedef struct tagLVKEYDOWN
 {
   NMHDR hdr;
   WORD  wVKey;
   UINT flags;
 } NMLVKEYDOWN, *LPNMLVKEYDOWN;
-#pragma pack(pop)
+#include <poppack.h>
 
 #define LV_KEYDOWN     NMLVKEYDOWN
 
@@ -4447,14 +4447,14 @@ typedef struct tagTCITEMW
 #define TCN_GETOBJECT           (TCN_FIRST - 3)
 #define TCN_FOCUSCHANGE         (TCN_FIRST - 4)
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 typedef struct tagTCKEYDOWN
 {
     NMHDR hdr;
     WORD wVKey;
     UINT flags;
 } NMTCKEYDOWN;
-#pragma pack(pop)
+#include <poppack.h>
 
 #define TC_KEYDOWN              NMTCKEYDOWN
 
@@ -5451,7 +5451,7 @@ static const WCHAR WC_SCROLLBARW[] = { 'S','c','r','o','l','l','B','a','r',0 };
 
 #ifndef NOTASKDIALOG
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 enum _TASKDIALOG_FLAGS
 {
@@ -5588,7 +5588,7 @@ WINCOMMCTRLAPI HRESULT WINAPI TaskDialog(HWND owner, HINSTANCE hinst, const WCHA
         const WCHAR *content, TASKDIALOG_COMMON_BUTTON_FLAGS common_buttons, const WCHAR *icon, int *button);
 WINCOMMCTRLAPI HRESULT WINAPI TaskDialogIndirect(const TASKDIALOGCONFIG *, int *, int *, BOOL *);
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* NOTASKDIALOG */
 
