diff --git a/include/commdlg.h b/include/commdlg.h
index 86961257350c349cedb7300f4435b5ad66847032..cd06a8c75a6d1654015435bfc94f02655f678a03 100644
--- a/include/commdlg.h
+++ b/include/commdlg.h
@@ -27,7 +27,7 @@ extern "C" {
 
 #include <prsht.h>
 #ifndef _WIN64
-#pragma pack(push,1)
+#include <pshpack1.h>
 #endif
 
 #ifndef WINCOMMDLGAPI
@@ -818,7 +818,7 @@ WINCOMMDLGAPI BOOL    WINAPI ChooseFontW(LPCHOOSEFONTW);
 #define ChooseFont    WINELIB_NAME_AW(ChooseFont)
 
 #ifndef _WIN64
-#pragma pack(pop)
+#include <poppack.h>
 #endif
 
 #ifdef __cplusplus
