diff --git a/include/cpl.h b/include/cpl.h
index 8939d421d83eaa123786465bd74615c94890b0f4..adabeba84843e3c3528b7a784106831aa5556525 100644
--- a/include/cpl.h
+++ b/include/cpl.h
@@ -21,7 +21,7 @@
 #ifndef _INC_CPL
 #define _INC_CPL
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -84,6 +84,6 @@ DECL_WINELIB_TYPE_AW(LPNEWCPLINFO)
 }
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* _INC_CPL */
