diff --git a/include/d2d1_1.idl b/include/d2d1_1.idl
index b9eb13e272aeba4b80d93ed1936982a727b27c5b..5560f24a38e6625ddcca2ce4133e8e53f046f687 100644
--- a/include/d2d1_1.idl
+++ b/include/d2d1_1.idl
@@ -29,14 +29,6 @@ interface ID2D1ColorContext;
 interface ID2D1Effect;
 interface ID2D1BitmapBrush1;
 
-typedef D2D_VECTOR_2F D2D1_VECTOR_2F;
-typedef D2D_VECTOR_3F D2D1_VECTOR_3F;
-typedef D2D_VECTOR_4F D2D1_VECTOR_4F;
-
-typedef D2D_MATRIX_4X3_F D2D1_MATRIX_4X3_F;
-typedef D2D_MATRIX_4X4_F D2D1_MATRIX_4X4_F;
-typedef D2D_MATRIX_5X4_F D2D1_MATRIX_5X4_F;
-
 cpp_quote("#ifndef __dwrite_h__")
 /* already defined in dwrite.h but needed for WIDL */
 typedef struct DWRITE_GLYPH_RUN_DESCRIPTION DWRITE_GLYPH_RUN_DESCRIPTION;
@@ -166,6 +158,7 @@ typedef enum D2D1_LAYER_OPTIONS1
 } D2D1_LAYER_OPTIONS1;
 
 struct D2D1_PROPERTY_BINDING;
+typedef D2D_MATRIX_4X4_F D2D1_MATRIX_4X4_F;
 
 typedef enum D2D1_PROPERTY_TYPE
 {
@@ -983,8 +976,6 @@ interface ID2D1Multithread : IUnknown
 
 [local] HRESULT __stdcall D2D1CreateDevice(IDXGIDevice *dxgi_device,
         const D2D1_CREATION_PROPERTIES *creation_properties, ID2D1Device **device);
-[local] HRESULT __stdcall D2D1CreateDeviceContext(IDXGISurface *dxgi_surface,
-        const D2D1_CREATION_PROPERTIES *creation_properties, ID2D1DeviceContext **context);
 [local] void __stdcall D2D1SinCos(float angle, float *s, float *c);
 [local] float __stdcall D2D1Tan(float angle);
 [local] float __stdcall D2D1Vec3Length(float x, float y, float z);
