diff --git a/include/d2d1_2.idl b/include/d2d1_2.idl
index c4cb06e4300da61e30ff010b210929de6afd753a..982131667df808eadac017f1f797d492f42c058b 100644
--- a/include/d2d1_2.idl
+++ b/include/d2d1_2.idl
@@ -101,5 +101,3 @@ interface ID2D1CommandSink1 : ID2D1CommandSink
         [in] D2D1_PRIMITIVE_BLEND primitive_blend
     );
 }
-
-[local] float __stdcall D2D1ComputeMaximumScaleFactor(const D2D1_MATRIX_3X2_F *matrix);
