diff --git a/include/d3d8types.h b/include/d3d8types.h
index 06c3b57dfa2b621fbdb207b599c1ef53a47599b5..163ee4fc3815434b78dfa07d9ca1d67c0a2bb432 100644
--- a/include/d3d8types.h
+++ b/include/d3d8types.h
@@ -20,7 +20,7 @@
 #define __WINE_D3D8TYPES_H
 
 #ifdef __i386__
-#pragma pack(push,4)
+#include <pshpack4.h>
 #endif
 
 /*****************************************************************************
@@ -150,9 +150,6 @@
 #define D3DCOLORWRITEENABLE_BLUE  (__MSABI_LONG(1)<<2)
 #define D3DCOLORWRITEENABLE_ALPHA (__MSABI_LONG(1)<<3)
 
-#define D3DDEVINFOID_RESOURCEMANAGER 5
-#define D3DDEVINFOID_VERTEXSTATS     6
-
 
 #ifndef MAKEFOURCC
 #define MAKEFOURCC(ch0, ch1, ch2, ch3)  \
@@ -1120,9 +1117,6 @@ typedef struct _D3DMATRIX {
     } DUMMYUNIONNAME;
 } D3DMATRIX;
 
-#define D3DPRESENT_RATE_DEFAULT 0
-#define D3DPRESENT_RATE_UNLIMITED 0x7fffffff
-
 typedef struct _D3DPRESENT_PARAMETERS_ {
     UINT                    BackBufferWidth;
     UINT                    BackBufferHeight;
@@ -1220,7 +1214,7 @@ typedef struct _D3DVOLUME_DESC {
 } D3DVOLUME_DESC;
 
 #ifdef __i386__
-#pragma pack(pop)
+#include <poppack.h>
 #endif
 
 #endif  /* __WINE_D3D8TYPES_H */
