diff --git a/include/d3d9caps.h b/include/d3d9caps.h
index ac21fac26128adcf1857c9a8588f942b78dce58a..c62780b4a0a78ff29874bd27545bd27f8ae995b7 100644
--- a/include/d3d9caps.h
+++ b/include/d3d9caps.h
@@ -21,7 +21,7 @@
 #define __WINE_D3D9CAPS_H
 
 #ifdef __i386__
-#pragma pack(push,4)
+#include <pshpack4.h>
 #endif
 
 /*
@@ -449,7 +449,7 @@ DEFINE_GUID(D3DKEYEXCHANGE_DXVA,        0x43d3775c, 0x38e5, 0x4924, 0x8d, 0x86,
 #endif
 
 #ifdef __i386__
-#pragma pack(pop)
+#include <poppack.h>
 #endif
 
 #endif
