diff --git a/include/d3dcompiler.h b/include/d3dcompiler.h
index 14e599081d8c403295f964773dcc866feeb72bfe..87821a9031ac550e06b5bdfc27e5025a0036927f 100644
--- a/include/d3dcompiler.h
+++ b/include/d3dcompiler.h
@@ -153,7 +153,6 @@ typedef HRESULT (WINAPI *pD3DPreprocess)(const void *data, SIZE_T size, const ch
         const D3D_SHADER_MACRO *defines, ID3DInclude *include,
         ID3DBlob **shader, ID3DBlob **error_messages);
 
-HRESULT WINAPI D3DCreateFunctionLinkingGraph(UINT flags, ID3D11FunctionLinkingGraph **graph);
 HRESULT WINAPI D3DCreateLinker(ID3D11Linker **linker);
 HRESULT WINAPI D3DLoadModule(const void *data, SIZE_T size, ID3D11Module **module);
 
