diff --git a/include/d3dx11.h b/include/d3dx11.h
index e1557857ffa5cf6d238f4e839b993457413bdaf1..7514328f4ce135aefe35ef954c73d18c2f61ebc5 100644
--- a/include/d3dx11.h
+++ b/include/d3dx11.h
@@ -28,8 +28,8 @@
 
 #include "d3d11.h"
 #include "d3dx11core.h"
-#include "d3dx11async.h"
 #include "d3dx11tex.h"
+#include "d3dx11async.h"
 
 #define _FACDD 0x876
 #define MAKE_DDHRESULT(code) MAKE_HRESULT(SEVERITY_ERROR, _FACDD, code)
