diff --git a/include/d3dx11async.h b/include/d3dx11async.h
index 2d85e4aa46f3fd036f8849d3992831436e91f6a2..e373b600a8bfc5669ef8179ada1ffb6b2ca90cbe 100644
--- a/include/d3dx11async.h
+++ b/include/d3dx11async.h
@@ -44,6 +44,12 @@ HRESULT WINAPI D3DX11CompileFromFileW(const WCHAR *filename, const D3D10_SHADER_
         ID3D10Include *include, const char *entry_point, const char *target, UINT sflags, UINT eflags,
         ID3DX11ThreadPump *pump, ID3D10Blob **shader, ID3D10Blob **error_messages, HRESULT *hresult);
 
+HRESULT WINAPI D3DX11CreateAsyncTextureProcessor(ID3D11Device *device, D3DX11_IMAGE_LOAD_INFO *info,
+        ID3DX11DataProcessor **processor);
+HRESULT WINAPI D3DX11CreateAsyncTextureInfoProcessor(D3DX11_IMAGE_INFO *info, ID3DX11DataProcessor **processor);
+HRESULT WINAPI D3DX11CreateAsyncShaderResourceViewProcessor(ID3D11Device *device, D3DX11_IMAGE_LOAD_INFO *info,
+        ID3DX11DataProcessor **processor);
+
 #ifdef __cplusplus
 }
 #endif
