diff --git a/include/d3dx11core.idl b/include/d3dx11core.idl
index f619e3b1b0291038d57c6f4e9f13902778fd6801..3b1339069eed540d14772ee5d7c44842c7e70ed5 100644
--- a/include/d3dx11core.idl
+++ b/include/d3dx11core.idl
@@ -63,3 +63,5 @@ interface ID3DX11ThreadPump : IUnknown
     HRESULT PurgeAllItems();
     HRESULT GetQueueStatus([in] UINT *io_queue, [in] UINT *process_queue, [in] UINT *device_queue);
 }
+
+cpp_quote("HRESULT WINAPI D3DX11CreateThreadPump(UINT io_threads, UINT proc_threads, ID3DX11ThreadPump **pump);")
