diff --git a/include/d3dx9tex.h b/include/d3dx9tex.h
index ae3fd8c34ff5bdec2337fab32eb2eb760e6ca40c..fa690107650cf3a35999164b41839e470254d042 100644
--- a/include/d3dx9tex.h
+++ b/include/d3dx9tex.h
@@ -181,8 +181,8 @@ HRESULT WINAPI D3DXLoadVolumeFromMemory(struct IDirect3DVolume9 *destvolume,
         D3DFORMAT srcformat, UINT srcrowpitch, UINT srcslicepitch, const PALETTEENTRY *srcpalette,
         const D3DBOX *srcbox, DWORD filter, D3DCOLOR colorkey);
 
-HRESULT WINAPI D3DXSaveVolumeToFileInMemory(struct ID3DXBuffer **dst_buffer, D3DXIMAGE_FILEFORMAT dst_format,
-        struct IDirect3DVolume9 *src_volume, const PALETTEENTRY *src_palette, const D3DBOX *src_box);
+HRESULT WINAPI D3DXSaveVolumeToFileInMemory(struct ID3DXBuffer **destbuffer, D3DXIMAGE_FILEFORMAT destformat,
+        struct IDirect3DVolume9 *srcvolume, const PALETTEENTRY *srcpalette, const D3DBOX *srcbox);
 
 HRESULT WINAPI D3DXSaveVolumeToFileA(const char *destfile, D3DXIMAGE_FILEFORMAT destformat,
         struct IDirect3DVolume9 *srcvolume, const PALETTEENTRY *srcpalette, const D3DBOX *srcbox);
