diff --git a/include/ddk/usb100.h b/include/ddk/usb100.h
index d8d3ca9947ec2fbe087de06f4feac4a6807169f5..7ec74933d0ff546b1535ed2605ecb6401f3682a7 100644
--- a/include/ddk/usb100.h
+++ b/include/ddk/usb100.h
@@ -28,7 +28,7 @@
 #define USB_CONFIG_POWER_DESCRIPTOR_TYPE          0x07
 #define USB_INTERFACE_POWER_DESCRIPTOR_TYPE       0x08
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 typedef struct _USB_DEVICE_DESCRIPTOR {
     UCHAR bLength;
@@ -113,6 +113,6 @@ typedef struct _USB_HUB_DESCRIPTOR {
 } USB_HUB_DESCRIPTOR;
 typedef struct _USB_HUB_DESCRIPTOR *PUSB_HUB_DESCRIPTOR;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif
