diff --git a/include/devpropdef.h b/include/devpropdef.h
index b654425dd91fe679805dd61ea53c6c5e3ad5b7c7..94046f5bce0ab90b89617aee3460c5b9270212e2 100644
--- a/include/devpropdef.h
+++ b/include/devpropdef.h
@@ -100,35 +100,4 @@ typedef struct _DEVPROPKEY {
 #else
 #define IsEqualDevPropKey(a,b) (((a).pid == (b).pid) && IsEqualIID(&(a).fmtid,&(b).fmtid))
 #endif
-
-typedef enum _DEVPROPSTORE
-{
-    DEVPROP_STORE_SYSTEM,
-    DEVPROP_STORE_USER,
-} DEVPROPSTORE, *PDEVPROPSTORE;
-
-typedef struct _DEVPROPCOMPKEY
-{
-    DEVPROPKEY Key;
-    DEVPROPSTORE Store;
-    PCWSTR LocaleName;
-} DEVPROPCOMPKEY, *PDEVPROPCOMPKEY;
-
-#ifndef IsEqualLocaleName
-#define IsEqualLocaleName(a,b) ((a) == (b) || ((a) && (b) && !wcsicmp((a),(b))))
-#endif
-
-#ifndef IsEqualDevPropCompKey
-#define IsEqualDevPropCompKey(a,b) (IsEqualDevPropKey((a).Key,(b).Key) && (a).Store == (b).Store && \
-                                    IsEqualLocaleName((a).LocaleName,(b).LocaleName))
-#endif
-
-typedef struct _DEVPROPERTY
-{
-    DEVPROPCOMPKEY CompKey;
-    DEVPROPTYPE Type;
-    ULONG BufferSize;
-    void *Buffer;
-} DEVPROPERTY, *PDEVPROPERTY;
-
 #endif
