diff --git a/include/digitalv.h b/include/digitalv.h
index 00c27445ebbe23b9c63192f16ad6f67a7d632eca..b2930fe602c61c5706deb6b6b26539ed0ef79fbc 100644
--- a/include/digitalv.h
+++ b/include/digitalv.h
@@ -23,7 +23,7 @@
  * Wine Digital Video extensions
  */
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -865,6 +865,6 @@ DECL_WINELIB_TYPE_AW(MCI_DGV_WINDOW_PARMS)
 }
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* __WINE_DIGITALV_H */
