diff --git a/include/dmksctrl.h b/include/dmksctrl.h
index 77caea452778e9a0cf2776976f44af39db3288f3..a35afe3f9d304ee9c0dcf7850ed8a81134d0cdb7 100644
--- a/include/dmksctrl.h
+++ b/include/dmksctrl.h
@@ -21,7 +21,7 @@
 #ifndef _DMKSCTRL_
 #define _DMKSCTRL_
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 #include <objbase.h>
 
@@ -95,7 +95,7 @@ DECLARE_INTERFACE_(IKsControl,IUnknown)
 
 #endif /* _KS_ */
 
-#pragma pack(pop)
+#include <poppack.h>
 
 
 DEFINE_GUID(IID_IKsControl, 0x28f54685, 0x06fd, 0x11d2, 0xb2, 0x7a, 0x00, 0xa0, 0xc9, 0x22, 0x31, 0x96);
