diff --git a/include/dmusicc.h b/include/dmusicc.h
index ff19c3ef6309c581074b59552c005e2ee059dbac..351f2733ee234ca8485d3de5c22ae7985cf934d1 100644
--- a/include/dmusicc.h
+++ b/include/dmusicc.h
@@ -33,7 +33,7 @@
 #include <dmusbuff.h>
 #include <strmif.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -723,6 +723,6 @@ DECLARE_INTERFACE_(IDirectMusicThru,IUnknown)
 }
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* __WINE_DMUSIC_CORE_H */
