diff --git a/include/dwmapi.h b/include/dwmapi.h
index 5381f6a3566d08373fd70cf2ed82cedc9d4f7c89..fb39aeeba1077a75354ab4103b4df0240c8f5b6e 100644
--- a/include/dwmapi.h
+++ b/include/dwmapi.h
@@ -32,7 +32,7 @@ extern "C" {
 DECLARE_HANDLE(HTHUMBNAIL);
 typedef HTHUMBNAIL *PHTHUMBNAIL;
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 static const UINT c_DwmMaxQueuedBuffers = 8;
 static const UINT c_DwmMaxMonitors = 16;
@@ -197,7 +197,7 @@ typedef struct _DWM_PRESENT_PARAMETERS {
     DWM_SOURCE_FRAME_SAMPLING eSampling;
 } DWM_PRESENT_PARAMETERS;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 DWMAPI DwmAttachMilContent(HWND);
 DWMAPI_(BOOL) DwmDefWindowProc(HWND, UINT, WPARAM, LPARAM, LRESULT*);
