diff --git a/include/fdi.h b/include/fdi.h
index dde4e468bc129733688d556c174cabfcb8910990..6659ce462ef02198b015fde3426fc098c86be54a 100644
--- a/include/fdi.h
+++ b/include/fdi.h
@@ -26,7 +26,7 @@ extern "C" {
 #endif /* defined(__cplusplus) */
 
 #ifndef _WIN64
-#pragma pack(push,4)
+#include <pshpack4.h>
 #endif
 
 #ifndef INCLUDED_TYPES_FCI_FDI
@@ -262,7 +262,7 @@ typedef INT_PTR (__cdecl *PFNFDINOTIFY)(FDINOTIFICATIONTYPE fdint,
 					   PFDINOTIFICATION pfdin)
 
 #ifndef _WIN64
-#pragma pack(push,1)
+#include <pshpack1.h>
 #endif
 
 typedef struct {
@@ -271,7 +271,7 @@ typedef struct {
 } FDISPILLFILE, *PFDISPILLFILE;
 
 #ifndef _WIN64
-#pragma pack(pop)
+#include <poppack.h>
 #endif
 
 #define cpuUNKNOWN (-1)  /* FDI does detection */
@@ -291,7 +291,7 @@ BOOL __cdecl FDITruncateCabinet(HFDI, char *, USHORT);
 /**********************************************************************/
 
 #ifndef _WIN64
-#pragma pack(pop)
+#include <poppack.h>
 #endif
 
 #ifdef __cplusplus
