diff --git a/include/gdiplus.h b/include/gdiplus.h
index 5f38ad0e648200700b9f04bbc23c3037d5a44435..e85343b7c17505d759a70d4f20a95099cbf9ad38 100644
--- a/include/gdiplus.h
+++ b/include/gdiplus.h
@@ -28,17 +28,16 @@ namespace Gdiplus
 #include "gdiplusmem.h"
     };
 
-#include "gdiplusbase.h"
 #include "gdiplusenums.h"
 #include "gdiplustypes.h"
 #include "gdiplusinit.h"
 #include "gdipluspixelformats.h"
-#include "gdipluscolor.h"
 #include "gdiplusmetaheader.h"
 #include "gdiplusimaging.h"
+#include "gdipluscolor.h"
 #include "gdipluscolormatrix.h"
-#include "gdipluseffects.h"
 #include "gdiplusgpstubs.h"
+#include "gdipluseffects.h"
 
     namespace DllExports
     {
