diff --git a/include/iads.idl b/include/iads.idl
index c6c72a287615b68db77411e92aa49dd715e6bcb7..6dfafce3a87de7d8c89f07926b51b4d3bc5549f3 100644
--- a/include/iads.idl
+++ b/include/iads.idl
@@ -152,20 +152,6 @@ typedef enum
     ADS_PASSWORD_ENCODE_CLEAR
 } ADS_PASSWORD_ENCODING_ENUM;
 
-typedef enum
-{
-    ADS_OPTION_SERVERNAME,
-    ADS_OPTION_REFERRALS,
-    ADS_OPTION_PAGE_SIZE,
-    ADS_OPTION_SECURITY_MASK,
-    ADS_OPTION_MUTUAL_AUTH_STATUS,
-    ADS_OPTION_QUOTA,
-    ADS_OPTION_PASSWORD_PORTNUMBER,
-    ADS_OPTION_PASSWORD_METHOD,
-    ADS_OPTION_ACCUMULATIVE_MODIFICATION,
-    ADS_OPTION_SKIP_SID_LOOKUP
-} ADS_OPTION_ENUM;
-
 typedef struct _ADS_CASEIGNORE_LIST
 {
     struct _ADS_CASEIGNORE_LIST *Next;
@@ -500,23 +486,6 @@ interface IADs : IDispatch
     HRESULT GetInfoEx( [in] VARIANT vProperties, [in] long lnReserved);
 }
 
-/*****************************************************************************
- *    IADsObjectOptions interface
- */
-[
-    odl,
-    uuid(46F14FDA-232B-11D1-A808-00C04FD8D5A8),
-    dual,
-    oleautomation
-]
-interface IADsObjectOptions : IDispatch
-{
-    [id(0x00000002)]
-    HRESULT GetOption([in] long option, [out, retval] VARIANT *var);
-    [id(0x00000003)]
-    HRESULT SetOption([in] long option, [in] VARIANT var);
-}
-
 /*****************************************************************************
  *    IADsMembers interface
  */
@@ -934,17 +903,6 @@ interface IADsPathname : IDispatch
     HRESULT put_EscapedMode(long mode);
 }
 
-[
-    helpstring("ADs LDAP Pathname Descriptor Object"),
-    uuid(080d0d78-f421-11d0-a36e-00c04fb950dc),
-    progid("Pathname"),
-    version(1.0)
-]
-coclass Pathname
-{
-    interface IADsPathname;
-}
-
 /*****************************************************************************
  *    IADsADSystemInfo interface
  */
@@ -995,68 +953,4 @@ coclass ADSystemInfo
     interface IADsADSystemInfo;
 }
 
-/*****************************************************************************
- *    IADsDNWithBinary interface
- */
-[
-    odl,
-    uuid(7e99c0a2-f935-11d2-ba96-00c04fb6d0d1),
-    dual,
-    oleautomation
-]
-interface IADsDNWithBinary : IDispatch
-{
-    [id(0x00000002), propget]
-    HRESULT BinaryValue([out, retval] VARIANT *value);
-    [id(0x00000002), propput]
-    HRESULT BinaryValue([in] VARIANT value);
-    [id(0x00000003), propget]
-    HRESULT DNString([out, retval] BSTR *value);
-    [id(0x00000003), propput]
-    HRESULT DNString([in] BSTR value);
-}
-
-[
-    helpstring("ADs DN With Binary Object"),
-    uuid(7e99c0a3-f935-11d2-ba96-00c04fb6d0d1),
-    progid("DNWithBinary"),
-    version(1.0)
-]
-coclass DNWithBinary
-{
-    interface IADsDNWithBinary;
-}
-
-/*****************************************************************************
- *    IADsDNWithString interface
- */
-[
-    odl,
-    uuid(370df02e-f934-11d2-ba96-00c04fb6d0d1),
-    dual,
-    oleautomation
-]
-interface IADsDNWithString : IDispatch
-{
-    [id(0x00000002), propget]
-    HRESULT StringValue([out, retval] BSTR *value);
-    [id(0x00000002), propput]
-    HRESULT StringValue([in] BSTR value);
-    [id(0x00000003), propget]
-    HRESULT DNString([out, retval] BSTR *value);
-    [id(0x00000003), propput]
-    HRESULT DNString([in] BSTR value);
-}
-
-[
-    helpstring("ADs DN With String Object"),
-    uuid(334857cc-f934-11d2-ba96-00c04fb6d0d1),
-    progid("DNWithString"),
-    version(1.0)
-]
-coclass DNWithString
-{
-    interface IADsDNWithString;
-}
-
 } /* library */
