diff --git a/include/minidumpapiset.h b/include/minidumpapiset.h
index 997c3567c69df9d6947e9fedabe838e3a0e19d43..c3cd43310aee67da8859071bf67c1f23b300c6a6 100644
--- a/include/minidumpapiset.h
+++ b/include/minidumpapiset.h
@@ -19,7 +19,7 @@
 #ifndef __WINE_MINIDUMP_H
 #define __WINE_MINIDUMP_H
 
-#pragma pack(push,4)
+#include <pshpack4.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -942,6 +942,6 @@ BOOL WINAPI MiniDumpReadDumpStream(PVOID base, ULONG index, PMINIDUMP_DIRECTORY
 }
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif  /* __WINE_MINIDUMP_H */
