diff --git a/include/mmreg.h b/include/mmreg.h
index 061aaa572a3d6e7e08211b57337a87e64155e590..a8b0056064e6133a6ebd60b94dde490c70e0c9dd 100644
--- a/include/mmreg.h
+++ b/include/mmreg.h
@@ -23,7 +23,7 @@
 #ifndef __WINE_MMREG_H
 #define __WINE_MMREG_H
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 /***********************************************************************
  * Defines/Enums
@@ -819,7 +819,7 @@ typedef struct tagEXBMINFOHEADER {
 
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* __WINE_MMREG_H */
 
