diff --git a/include/msacm.h b/include/msacm.h
index 4f3060cec2fc9276d9bea229d7f05e8baf5de94f..8d459cb24721f20e30ac9770962f1ae1b045f728 100644
--- a/include/msacm.h
+++ b/include/msacm.h
@@ -25,7 +25,7 @@
 extern "C" {
 #endif /* defined(__cplusplus) */
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 #define ACMAPI              WINAPI
 
@@ -774,7 +774,7 @@ MMRESULT WINAPI acmStreamUnprepareHeader(
   HACMSTREAM has, PACMSTREAMHEADER pash, DWORD fdwUnprepare
 );
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #ifdef __cplusplus
 } /* extern "C" */
