diff --git a/include/mscat.h b/include/mscat.h
index 1a82bbe1b82175758d78ce9eb70bd553773ef814..366dbbb0d1ee9f8f0bc13099e70ff7635bd26d17 100644
--- a/include/mscat.h
+++ b/include/mscat.h
@@ -66,7 +66,7 @@ extern "C" {
 #define CRYPTCAT_VERSION_1  0x100
 #define CRYPTCAT_VERSION_2  0x200
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 typedef struct CRYPTCATATTRIBUTE_
 {
@@ -110,7 +110,7 @@ typedef struct CRYPTCATCDF_
     HANDLE hCATStore;
 } CRYPTCATCDF;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef void (WINAPI *PFN_CDF_PARSE_ERROR_CALLBACK)(DWORD, DWORD, WCHAR *);
 
