diff --git a/include/msdelta.h b/include/msdelta.h
index d674ec19bef57c578f23e8c5f4c7bf7fe8354b76..aa7161c6b0ac363134bc95e57d91ff47db11a74f 100644
--- a/include/msdelta.h
+++ b/include/msdelta.h
@@ -26,8 +26,8 @@ extern "C" {
 #include <windows.h>
 #include <wincrypt.h>
 
-typedef INT64 DELTA_FILE_TYPE;
-typedef INT64 DELTA_FLAG_TYPE;
+typedef __int64 DELTA_FILE_TYPE;
+typedef __int64 DELTA_FLAG_TYPE;
 
 #define DELTA_FILE_SIZE_LIMIT         (32*1024*1024)
 #define DELTA_OPTIONS_SIZE_LIMIT      (128*1024*1024)
