diff --git a/include/mshtmhst.idl b/include/mshtmhst.idl
index 6cc500ed7a9eaef1476066ac8ef852ccb25e9cbd..efc51c6a5da24bf607138b16a697530a92036db3 100644
--- a/include/mshtmhst.idl
+++ b/include/mshtmhst.idl
@@ -163,7 +163,7 @@ typedef enum tagDOCHOSTUIFLAG
     object,
     uuid(BD3F23C0-D43E-11CF-893B-00AA00BDCE1A),
     pointer_default(unique),
-    local
+ /* local, CXHACK 15579 */
 ]
 interface IDocHostUIHandler : IUnknown
 {
@@ -237,7 +237,7 @@ cpp_quote("DEFINE_GUID(CGID_DocHostCommandHandler,0xf38bc242,0xb950,0x11d1,0x89,
     object,
     uuid(3050F6D0-98b5-11CF-BB82-00AA00BDCE0B),
     pointer_default(unique),
-    local
+ /* local, CXHACK 15579 */
 ]
 interface IDocHostUIHandler2 : IDocHostUIHandler
 {
@@ -253,7 +253,7 @@ interface IDocHostUIHandler2 : IDocHostUIHandler
     object,
     uuid(3050f3f0-98b5-11cf-bb82-00aa00bdce0b),
     pointer_default(unique),
-    local
+ /* local, CXHACK 15579 */
 ]
 interface ICustomDoc : IUnknown
 {
