diff --git a/include/mshtml.idl b/include/mshtml.idl
index 1d2896f88b8cd0b13a72a72308283d6aa1c7dc84..7fb245f5d5c25eb1781f25ce4519e4a69f4da402 100644
--- a/include/mshtml.idl
+++ b/include/mshtml.idl
@@ -8408,36 +8408,6 @@ interface IHTMLDOMAttribute2 : IDispatch
             [retval, out] IHTMLDOMAttribute **clonedNode);
 }
 
-/*****************************************************************************
- *    IHTMLDOMAttribute3 interface
- */
-[
-    odl,
-    oleautomation,
-    dual,
-    uuid(30510468-98b5-11cf-bb82-00aa00bdce0b)
-]
-interface IHTMLDOMAttribute3 : IDispatch
-{
-    [propput, id(DISPID_IHTMLDOMATTRIBUTE3_IE8_NODEVALUE)]
-    HRESULT nodeValue([in] VARIANT v);
-
-    [propget, id(DISPID_IHTMLDOMATTRIBUTE3_IE8_NODEVALUE)]
-    HRESULT nodeValue([retval, out] VARIANT *p);
-
-    [propput, id(DISPID_IHTMLDOMATTRIBUTE3_IE8_VALUE)]
-    HRESULT value([in] BSTR v);
-
-    [propget, id(DISPID_IHTMLDOMATTRIBUTE3_IE8_VALUE)]
-    HRESULT value([retval, out] BSTR *p);
-
-    [propget, id(DISPID_IHTMLDOMATTRIBUTE3_IE8_SPECIFIED)]
-    HRESULT specified([retval, out] VARIANT_BOOL *p);
-
-    [propget, id(DISPID_IHTMLDOMATTRIBUTE3_OWNERELEMENT)]
-    HRESULT ownerElement([retval, out] IHTMLElement2 **p);
-}
-
 /*****************************************************************************
  *    IHTMLDOMTextNode interface
  */
@@ -8958,58 +8928,6 @@ interface IHTMLAttributeCollection3 : IDispatch
     HRESULT length([retval, out] long *p);
 }
 
-/*****************************************************************************
- *    IHTMLAttributeCollection4 interface
- */
-[
-    odl,
-    oleautomation,
-    dual,
-    uuid(305106fa-98b5-11cf-bb82-00aa00bdce0b)
-]
-interface IHTMLAttributeCollection4 : IDispatch
-{
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_GETNAMEDITEMNS)]
-    HRESULT getNamedItemNS(
-            [in] VARIANT *pvarNS,
-            [in] BSTR bstrName,
-            [out, retval] IHTMLDOMAttribute2 **ppNodeOut);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_SETNAMEDITEMNS)]
-    HRESULT setNamedItemNS(
-            [in] IHTMLDOMAttribute2 *pNodeIn,
-            [out, retval] IHTMLDOMAttribute2 **ppNodeOut);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_REMOVENAMEDITEMNS)]
-    HRESULT removeNamedItemNS(
-            [in] VARIANT *pvarNS,
-            [in] BSTR bstrName,
-            [out, retval] IHTMLDOMAttribute2 **ppNodeOut);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_GETNAMEDITEM)]
-    HRESULT getNamedItem(
-            [in] BSTR bstrName,
-            [out, retval] IHTMLDOMAttribute2 **ppNodeOut);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_SETNAMEDITEM)]
-    HRESULT setNamedItem(
-            [in] IHTMLDOMAttribute2 *pNodeIn,
-            [out, retval] IHTMLDOMAttribute2 **ppNodeOut);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_REMOVENAMEDITEM)]
-    HRESULT removeNamedItem(
-            [in] BSTR bstrName,
-            [out, retval] IHTMLDOMAttribute2 **ppNodeOut);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_ITEM)]
-    HRESULT item(
-            [in] long index,
-            [out, retval] IHTMLDOMAttribute2 **ppNodeOut);
-
-    [propget, id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_LENGTH)]
-    HRESULT length([out, retval] long *p);
-}
-
 /*****************************************************************************
  *    DispHTMLAttributeCollection dispinterface
  */
@@ -9054,30 +8972,6 @@ methods:
     [propget, id(DISPID_IHTMLATTRIBUTECOLLECTION3_IE8_LENGTH)]
     long ie8_length();
 
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_GETNAMEDITEMNS)]
-    IHTMLDOMAttribute2 *getNamedItemNS([in] VARIANT *pvarNS, [in] BSTR bstrName);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_SETNAMEDITEMNS)]
-    IHTMLDOMAttribute2 *setNamedItemNS([in] IHTMLDOMAttribute2 *pNodeIn);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_REMOVENAMEDITEMNS)]
-    IHTMLDOMAttribute2 *removeNamedItemNS([in] VARIANT *pvarNS, [in] BSTR bstrName);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_GETNAMEDITEM)]
-    IHTMLDOMAttribute2 *ie9_getNamedItem([in] BSTR bstrName);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_SETNAMEDITEM)]
-    IHTMLDOMAttribute2 *ie9_setNamedItem([in] IHTMLDOMAttribute2 *pNodeIn);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_REMOVENAMEDITEM)]
-    IHTMLDOMAttribute2 *ie9_removeNamedItem([in] BSTR bstrName);
-
-    [id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_ITEM)]
-    IHTMLDOMAttribute2 *ie9_item([in] long index);
-
-    [propget, id(DISPID_IHTMLATTRIBUTECOLLECTION4_IE9_LENGTH)]
-    long ie9_length();
-
     [propget, id(DISPID_IHTMLDOMCONSTRUCTOR_CONSTRUCTOR), hidden]
     IDispatch *constructor();
 }
@@ -16332,70 +16226,6 @@ methods:
     void setSelectionRange([in] long start, [in] long end);
 }
 
-[
-    odl,
-    oleautomation,
-    dual,
-    uuid(3050f2bc-98b5-11cf-bb82-00aa00bdce0b)
-]
-interface IHTMLOptionButtonElement : IDispatch
-{
-    [propput, id(DISPID_IHTMLOPTIONBUTTONELEMENT_VALUE), displaybind, bindable]
-    HRESULT value([in] BSTR v);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_VALUE), displaybind, bindable]
-    HRESULT value([retval, out] BSTR *p);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_TYPE)]
-    HRESULT type([retval, out] BSTR *p);
-
-    [propput, id(DISPID_IHTMLOPTIONBUTTONELEMENT_NAME), displaybind, bindable]
-    HRESULT name([in] BSTR v);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_NAME), displaybind, bindable]
-    HRESULT name([retval, out] BSTR *p);
-
-    [propput, id(DISPID_IHTMLOPTIONBUTTONELEMENT_CHECKED), bindable]
-    HRESULT checked([in] VARIANT_BOOL v);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_CHECKED), bindable]
-    HRESULT checked([retval, out] VARIANT_BOOL *p);
-
-    [propput, id(DISPID_IHTMLOPTIONBUTTONELEMENT_DEFAULTCHECKED), bindable]
-    HRESULT defaultChecked([in] VARIANT_BOOL v);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_DEFAULTCHECKED), bindable]
-    HRESULT defaultChecked([retval, out] VARIANT_BOOL *p);
-
-    [propput, id(DISPID_IHTMLOPTIONBUTTONELEMENT_ONCHANGE), displaybind, bindable]
-    HRESULT onchange([in] VARIANT v);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_ONCHANGE), displaybind, bindable]
-    HRESULT onchange([retval, out] VARIANT *p);
-
-    [propput, id(DISPID_IHTMLOPTIONBUTTONELEMENT_DISABLED), displaybind, bindable]
-    HRESULT disabled([in] VARIANT_BOOL v);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_DISABLED), displaybind, bindable]
-    HRESULT disabled([retval, out] VARIANT_BOOL *p);
-
-    [propput, id(DISPID_IHTMLOPTIONBUTTONELEMENT_STATUS)]
-    HRESULT status([in] VARIANT_BOOL v);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_STATUS)]
-    HRESULT status([retval, out] VARIANT_BOOL *p);
-
-    [propput, id(DISPID_IHTMLOPTIONBUTTONELEMENT_INDETERMINATE), bindable]
-    HRESULT indeterminate([in] VARIANT_BOOL v);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_INDETERMINATE), bindable]
-    HRESULT indeterminate([retval, out] VARIANT_BOOL *p);
-
-    [propget, id(DISPID_IHTMLOPTIONBUTTONELEMENT_FORM)]
-    HRESULT form([retval, out] IHTMLFormElement* *p);
-};
-
-
 /*****************************************************************************
  *    HTMLInputElement coclass
  */
@@ -16419,7 +16249,7 @@ coclass HTMLInputElement
  /* interface IHTMLInputHiddenElement; */
  /* interface IHTMLInputButtonElement; */
  /* interface IHTMLInputFileElement; */
-    interface IHTMLOptionButtonElement;
+ /* interface IHTMLOptionButtonElement; */
  /* interface IHTMLInputImage; */
  /* interface IHTMLInputElement3; */
  /* interface IHTMLInputRangeElement; */
@@ -30523,7 +30353,11 @@ interface IElementBehaviorFactory : IUnknown
 
 } /* library MSHTML */
 
+#ifdef MSHTML_MARSHALING_HACK
+#define IOleCommandTarget IUnknown
+#else
 interface IOleCommandTarget;
+#endif
 
 /*****************************************************************************
  *    IHTMLPrivateWindow interface
@@ -30531,7 +30365,7 @@ interface IOleCommandTarget;
 [
     object,
     uuid(3050f6dc-98b5-11cf-bb82-00aa00bdce0b),
-    local
+ /* local, CXHACK 15579 */
 ]
 interface IHTMLPrivateWindow : IUnknown
 {
@@ -30540,7 +30374,7 @@ interface IHTMLPrivateWindow : IUnknown
     HRESULT SetPICSTarget(IOleCommandTarget *cmdtrg);
     HRESULT PICSComplete(int arg);
     HRESULT FindWindowByName(LPCWSTR name, IHTMLWindow2 **ret);
-    HRESULT GetAddressBarUrl(BSTR *url);
+    HRESULT GetAddressBarUrl([out, retval] BSTR *url);
 }
 
 /*****************************************************************************
@@ -30549,7 +30383,7 @@ interface IHTMLPrivateWindow : IUnknown
 [
     object,
     uuid(3050f804-98b5-11cf-bb82-00aa00bdce0b),
-    local
+ /* local, CXHACK 15579 */
 ]
 interface IWebBrowserPriv : IUnknown
 {
@@ -30564,7 +30398,7 @@ interface IWebBrowserPriv : IUnknown
 [
     object,
     uuid(3ed72303-6ffc-4214-ba90-faf1862dec8a),
-    local
+ /* local, CXHACK 15579 */
 ]
 interface IWebBrowserPriv2IE8 : IUnknown
 {
@@ -30586,7 +30420,7 @@ interface IWebBrowserPriv2IE8 : IUnknown
 [
     object,
     uuid(3ed72303-6ffc-4214-ba90-faf1862dec8a),
-    local
+ /* local, CXHACK 15579 */
 ]
 interface IWebBrowserPriv2IE9 : IUnknown
 {
@@ -30594,3 +30428,7 @@ interface IWebBrowserPriv2IE9 : IUnknown
             VARIANT *headers, IBindCtx *bind_ctx, LPOLESTR url_fragment, DWORD unused);
     /* Probably more */
 }
+
+#ifdef MSHTML_MARSHALING_HACK
+#undef IOleCommandTarget
+#endif
