diff --git a/include/mssip.h b/include/mssip.h
index c74b7e40f209ac9b42075bc49fc7d1e21f8e52b1..d002f9612332e827d85cef26c3c6c6ac404ad757 100644
--- a/include/mssip.h
+++ b/include/mssip.h
@@ -48,7 +48,7 @@ typedef CRYPT_HASH_BLOB CRYPT_DIGEST_DATA;
 
 /**********************************************************************/
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct SIP_SUBJECTINFO_ {
     DWORD cbSize;
     GUID *pgSubjectType;
@@ -77,26 +77,26 @@ typedef struct SIP_SUBJECTINFO_ {
 
     LPVOID pClientData;
 } SIP_SUBJECTINFO, *LPSIP_SUBJECTINFO;
-#pragma pack(pop)
+#include <poppack.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct MS_ADDINFO_FLAT_ {
   DWORD cbStruct;
 
   struct SIP_INDIRECT_DATA_ *pIndirectData;
 } MS_ADDINFO_FLAT, *PMS_ADDINFO_FLAT;
-#pragma pack(pop)
+#include <poppack.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct MS_ADDINFO_CATALOGMEMBER_ {
   DWORD cbStruct;
 
   struct CRYPTCATSTORE_  *pStore;
   struct CRYPTCATMEMBER_ *pMember;
 } MS_ADDINFO_CATALOGMEMBER, *PMS_ADDINFO_CATALOGMEMBER;
-#pragma pack(pop)
+#include <poppack.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct MS_ADDINFO_BLOB_ {
   DWORD cbStruct;
 
@@ -106,15 +106,15 @@ typedef struct MS_ADDINFO_BLOB_ {
   DWORD cbMemSignedMsg;
   BYTE *pbMemSignedMsg;
 } MS_ADDINFO_BLOB, *PMS_ADDINFO_BLOB;
-#pragma pack(pop)
+#include <poppack.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct SIP_INDIRECT_DATA_ {
   CRYPT_ATTRIBUTE_TYPE_VALUE Data;
   CRYPT_ALGORITHM_IDENTIFIER DigestAlgorithm;
   CRYPT_HASH_BLOB            Digest;
 } SIP_INDIRECT_DATA, *PSIP_INDIRECT_DATA;
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef BOOL (WINAPI * pCryptSIPGetSignedDataMsg)(SIP_SUBJECTINFO *,DWORD *,DWORD,DWORD *,BYTE *);
 typedef BOOL (WINAPI * pCryptSIPPutSignedDataMsg)(SIP_SUBJECTINFO *,DWORD,DWORD *,DWORD,BYTE *);
@@ -122,7 +122,7 @@ typedef BOOL (WINAPI * pCryptSIPCreateIndirectData)(SIP_SUBJECTINFO *,DWORD *,SI
 typedef BOOL (WINAPI * pCryptSIPVerifyIndirectData)(SIP_SUBJECTINFO *,SIP_INDIRECT_DATA *);
 typedef BOOL (WINAPI * pCryptSIPRemoveSignedDataMsg)(SIP_SUBJECTINFO *,DWORD);
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct SIP_DISPATCH_INFO_ {
   DWORD cbSize;
 
@@ -134,12 +134,12 @@ typedef struct SIP_DISPATCH_INFO_ {
   pCryptSIPVerifyIndirectData  pfVerify;
   pCryptSIPRemoveSignedDataMsg pfRemove;
 } SIP_DISPATCH_INFO, *LPSIP_DISPATCH_INFO;
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef BOOL (WINAPI *pfnIsFileSupported)(HANDLE,GUID *);
 typedef BOOL (WINAPI *pfnIsFileSupportedName)(WCHAR *,GUID *);
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct SIP_ADD_NEWPROVIDER_
 {
   DWORD cbStruct;
@@ -161,7 +161,7 @@ typedef struct SIP_ADD_NEWPROVIDER_
 
   WCHAR *pwszGetCapFuncName;
 } SIP_ADD_NEWPROVIDER, *PSIP_ADD_NEWPROVIDER;
-#pragma pack(pop)
+#include <poppack.h>
 
 /**********************************************************************/
 
