diff --git a/include/msvcrt/complex.h b/include/msvcrt/complex.h
index 3d27f4429cb3c2cd0d646e81120aa353adeb97d6..2d7009c470c1c61b496a97af23264bb3c90fc57f 100644
--- a/include/msvcrt/complex.h
+++ b/include/msvcrt/complex.h
@@ -24,35 +24,4 @@ typedef struct _C_float_complex
 typedef _C_double_complex _Dcomplex;
 typedef _C_float_complex _Fcomplex;
 
-_ACRTIMP _Dcomplex __cdecl _Cbuild(double, double);
-_ACRTIMP _Dcomplex __cdecl cexp(_Dcomplex);
-
-_ACRTIMP double __cdecl carg(_Dcomplex);
-_ACRTIMP double __cdecl cimag(_Dcomplex);
-_ACRTIMP double __cdecl creal(_Dcomplex);
-
-#if defined(__i386__) && !defined(__MINGW32__) && !defined(_MSC_VER)
-/* Note: this should return a _Fcomplex, but calling convention for returning
- * structures is different between Windows and gcc on i386. */
-_ACRTIMP unsigned __int64 __cdecl _FCbuild(float, float);
-
-static inline _Fcomplex __cdecl __wine__FCbuild(float r, float i)
-{
-    union {
-        _Fcomplex c;
-        unsigned __int64 ull;
-    } u;
-    u.ull = _FCbuild(r, i);
-    return u.c;
-}
-#define _FCbuild(r, i) __wine__FCbuild(r, i)
-
-#else
-_ACRTIMP _Fcomplex __cdecl _FCbuild(float, float);
-#endif
-
-_ACRTIMP float __cdecl cargf(_Fcomplex);
-_ACRTIMP float __cdecl cimagf(_Fcomplex);
-_ACRTIMP float __cdecl crealf(_Fcomplex);
-
 #endif /* _COMPLEX_H_DEFINED */
