diff --git a/include/msvcrt/corecrt_startup.h b/include/msvcrt/corecrt_startup.h
index c1c519da6180e2a75f4ec328fae6454405a20af0..58450b94ca0de9f348cbc09d11813001727b2b21 100644
--- a/include/msvcrt/corecrt_startup.h
+++ b/include/msvcrt/corecrt_startup.h
@@ -26,8 +26,6 @@ typedef void (__cdecl *_PVFV)(void);
 typedef int (__cdecl *_PIFV)(void);
 typedef void (__cdecl *_PVFI)(int);
 
-_ACRTIMP int __cdecl _initterm_e(_PIFV *, _PIFV *);
-
 typedef struct _onexit_table_t {
     _PVFV *_first;
     _PVFV *_last;
