diff --git a/include/msvcrt/corecrt_wstdio.h b/include/msvcrt/corecrt_wstdio.h
index ca5d804e30b5c9b8867c8e5d1259dd541d6c2213..27e29700d0198b758a296b4122e5236bf4a7c546 100644
--- a/include/msvcrt/corecrt_wstdio.h
+++ b/include/msvcrt/corecrt_wstdio.h
@@ -18,11 +18,11 @@
 extern "C" {
 #endif
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 #ifndef _FILE_DEFINED
 #define _FILE_DEFINED
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct _iobuf
 {
 #ifdef _UCRT
@@ -38,7 +38,7 @@ typedef struct _iobuf
   char* _tmpfname;
 #endif
 } FILE;
-#pragma pack(pop)
+#include <poppack.h>
 #endif  /* _FILE_DEFINED */
 
 #ifndef WEOF
@@ -412,6 +412,6 @@ _ACRTIMP int __cdecl wscanf_s(const wchar_t*,...);
 }
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* _WSTDIO_DEFINED */
