diff --git a/include/msvcrt/direct.h b/include/msvcrt/direct.h
index 9ff3e23cdd0d8180d5a182ee88b4b411acc6d083..83ac36899894fad558d7dc3e21341cb71d6da48a 100644
--- a/include/msvcrt/direct.h
+++ b/include/msvcrt/direct.h
@@ -10,7 +10,7 @@
 
 #include <corecrt_wdirect.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -45,6 +45,6 @@ static inline char* getcwd(char * buf, int size) { return _getcwd(buf, size); }
 static inline int mkdir(const char* newdir) { return _mkdir(newdir); }
 static inline int rmdir(const char* dir) { return _rmdir(dir); }
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* __WINE_DIRECT_H */
