diff --git a/include/msvcrt/sys/stat.h b/include/msvcrt/sys/stat.h
index 17c595542a1ee9a57dde1820d5f1daff41243553..6e5a400de6c60eec7febade397d5201af8103961 100644
--- a/include/msvcrt/sys/stat.h
+++ b/include/msvcrt/sys/stat.h
@@ -11,7 +11,7 @@
 #include <corecrt.h>
 #include <sys/types.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 #ifndef _DEV_T_DEFINED
 # ifdef _CRTDLL
@@ -222,6 +222,6 @@ static inline int umask(int fd) { return _umask(fd); }
 #define _UMASK_DEFINED
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* __WINE_SYS_STAT_H */
