diff --git a/include/msvcrt/sys/timeb.h b/include/msvcrt/sys/timeb.h
index 357ff77cff841567ab6c63ac5e2e4584500a7c18..e948daab012ae97b9f6d7a4d0f0022376bcb9e8e 100644
--- a/include/msvcrt/sys/timeb.h
+++ b/include/msvcrt/sys/timeb.h
@@ -22,7 +22,7 @@
 
 #include <corecrt.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 #ifndef _TIMEB_DEFINED
 #define _TIMEB_DEFINED
@@ -71,6 +71,6 @@ static inline void __cdecl _ftime(struct _timeb *tb) { return _ftime64((struct _
 
 static inline void ftime(struct _timeb* ptr) { return _ftime(ptr); }
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* __WINE_SYS_TIMEB_H */
