diff --git a/include/msvcrt/time.h b/include/msvcrt/time.h
index 04d8a4b22497634df9280a7c19eac22c01dac6f6..05965c4dad221e28ef300e0cfe8fc89390bc0468 100644
--- a/include/msvcrt/time.h
+++ b/include/msvcrt/time.h
@@ -22,7 +22,7 @@
 
 #include <corecrt_wtime.h>
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 #ifndef _CLOCK_T_DEFINED
 typedef __msvcrt_long clock_t;
@@ -72,7 +72,7 @@ extern __msvcrt_long _timezone;
 extern char *_tzname;
 #endif
 
-#if _MSVCR_VER < 120 && defined(_USE_32BIT_TIME_T)
+#if !defined(_UCRT) && defined(_USE_32BIT_TIME_T)
 #define _ctime32     ctime
 #define _difftime32  difftime
 #define _gmtime32    gmtime
@@ -137,6 +137,6 @@ static inline time_t time(time_t *t) { return _time32(t); }
 }
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* __WINE_TIME_H */
