diff --git a/include/msxml.idl b/include/msxml.idl
index 283a0f48bb6ffd0e04073bb0842d90cb7871f5c1..375bb64cd7c11079adbebe998a4c890c7b4d439a 100644
--- a/include/msxml.idl
+++ b/include/msxml.idl
@@ -164,7 +164,7 @@ interface IXMLDocument : IDispatch
     HRESULT doctype( [retval, out] BSTR *p );
 
     [propget, id(DISPID_XMLDOCUMENT_DTDURL)]
-    HRESULT dtdURL( [retval, out] BSTR *p );
+    HRESULT dtdURl( [retval, out] BSTR *p );
 
     [id(DISPID_XMLDOCUMENT_CREATEELEMENT)]
     HRESULT createElement(
