diff --git a/include/msxml6.idl b/include/msxml6.idl
index ead87be7516e893e44daa622079cfc2584eb0485..a02f7b2db143c7c6a93a57498d27cd5386a7bafd 100644
--- a/include/msxml6.idl
+++ b/include/msxml6.idl
@@ -1715,17 +1715,6 @@ interface ISAXDeclHandler : IUnknown
         [in] int nSystemId);
 }
 
-[
-    helpstring("Free Threaded XML HTTP Request class 6.0"),
-    progid("Msxml2.FreeThreadedXMLHTTP60.6.0"),
-    threading(both),
-    uuid(88d96a09-f192-11d4-a65f-0040963251e5)
-]
-coclass FreeThreadedXMLHTTP60
-{
-    [default] interface IXMLHTTPRequest2;
-}
-
 [
     object,
     local,
@@ -3054,24 +3043,23 @@ interface ISchemaNotation;
 } __msxml6_ReferenceRemainingTypes__;
 
 [
-    helpstring("XML DOM Document 6.0"),
-    progid("Msxml2.DOMDocument.6.0"),
+    helpstring("Free Threaded XML HTTP Request class 6.0"),
+    progid("Msxml2.FreeThreadedXMLHTTP60.6.0"),
     threading(both),
-    uuid(88d96a05-f192-11d4-a65f-0040963251e5)
+    uuid(88d96a09-f192-11d4-a65f-0040963251e5)
 ]
-coclass DOMDocument60
+coclass FreeThreadedXMLHTTP60
 {
-    [default] interface IXMLDOMDocument3;
-    [default, source] dispinterface XMLDOMDocumentEvents;
+    [default] interface IXMLHTTPRequest2;
 }
 
 [
-    helpstring("Free threaded XML DOM Document 6.0"),
-    progid("Msxml2.FreeThreadedDOMDocument.6.0"),
+    helpstring("XML DOM Document 6.0"),
+    progid("Msxml2.DOMDocument.6.0"),
     threading(both),
-    uuid(88d96a06-f192-11d4-a65f-0040963251e5),
+    uuid(88d96a05-f192-11d4-a65f-0040963251e5)
 ]
-coclass FreeThreadedDOMDocument60
+coclass DOMDocument60
 {
     [default] interface IXMLDOMDocument3;
     [default, source] dispinterface XMLDOMDocumentEvents;
@@ -3182,6 +3170,18 @@ coclass XSLTemplate60
     [default] interface IXSLTemplate;
 }
 
+[
+    helpstring("Free threaded XML DOM Document 6.0"),
+    progid("Msxml2.FreeThreadedDOMDocument.6.0"),
+    threading(both),
+    uuid(88d96a06-f192-11d4-a65f-0040963251e5),
+]
+coclass FreeThreadedDOMDocument60
+{
+    [default] interface IXMLDOMDocument3;
+    [default, source] dispinterface XMLDOMDocumentEvents;
+}
+
 [
     helpstring("XML HTTP 6.0"),
     progid("Msxml2.XMLHTTP.6.0"),
