diff --git a/include/ntdef.h b/include/ntdef.h
index f176bc201102881737f24b696e5782fc4987cbc0..e840bd3f9fe974f4df898a727c9fbeada9cbd60c 100644
--- a/include/ntdef.h
+++ b/include/ntdef.h
@@ -51,30 +51,17 @@ typedef enum _WAIT_TYPE {
 #define NT_WARNING(status)      ((((NTSTATUS)(status)) & 0xc0000000) == 0x80000000)
 #define NT_ERROR(status)        ((((NTSTATUS)(status)) & 0xc0000000) == 0xc0000000)
 
-#define InitializeObjectAttributes(p,n,a,r,s) \
-    do { \
-        (p)->Length = sizeof(OBJECT_ATTRIBUTES); \
-        (p)->RootDirectory = r; \
-        (p)->Attributes = a; \
-        (p)->ObjectName = n; \
-        (p)->SecurityDescriptor = s; \
-        (p)->SecurityQualityOfService = NULL; \
-    } while (0)
-
 #ifndef BASETYPES
 #define BASETYPES
 typedef unsigned char UCHAR, *PUCHAR;
 typedef unsigned short USHORT, *PUSHORT;
-#if !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES) && !defined(WINE_UNIX_LIB)
+#if !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES)
 typedef unsigned long ULONG, *PULONG;
 #else
 typedef unsigned int ULONG, *PULONG;
 #endif
 #endif
 
-typedef ULONG CLONG;
-typedef CLONG *PCLONG;
-
 typedef struct _RTL_BALANCED_NODE
 {
     union
