diff --git a/include/ntquery.h b/include/ntquery.h
index 0eb99aa9c1141191c69fdedb2ee7eea3fe2af015..307304fe5ad795d9557d9f9da2a19d4d1d576f7f 100644
--- a/include/ntquery.h
+++ b/include/ntquery.h
@@ -20,7 +20,7 @@
 #define __WINE_NTQUERY_H__
 
 #include "stgprop.h"
-#pragma pack(push,4)
+#include <pshpack4.h>
 
 typedef struct _CI_STATE
 {
@@ -41,7 +41,7 @@ typedef struct _CI_STATE
     DWORD dwPropCacheSize;
 } CI_STATE;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #ifdef __cplusplus
 extern "C" {
