diff --git a/include/oledb.idl b/include/oledb.idl
index ff3512f3b26af3062b5e7a561e0ae2b1a15e6cc4..f123e3e3c9d6ae4d4b052761b9e00692ad8ee6fc 100644
--- a/include/oledb.idl
+++ b/include/oledb.idl
@@ -16,9 +16,9 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 cpp_quote("#ifdef _WIN64")
-cpp_quote("#pragma pack(push,8)")
+cpp_quote("#include <pshpack8.h>")
 cpp_quote("#else")
-cpp_quote("#pragma pack(push,2)")
+cpp_quote("#include <pshpack2.h>")
 cpp_quote("#endif")
 cpp_quote("")
 
@@ -52,9 +52,6 @@ typedef LONG_PTR DB_LRESERVE;
 #include "rstxsc.idl"
 #include "chprst.idl"
 #include "rstfnd.idl"
-#include "rstvw.idl"
-#include "vwchp.idl"
-#include "vwflt.idl"
 #include "cmdbas.idl"
 #include "cmdtxt.idl"
 #include "colinf.idl"
@@ -70,7 +67,6 @@ typedef LONG_PTR DB_LRESERVE;
 #include "sesprp.idl"
 #include "rstnot.idl"
 #include "opnrst.idl"
-#include "dbsrst.idl"
 #include "row.idl"
 #include "rowchg.idl"
 #include "rowpos.idl"
@@ -78,7 +74,6 @@ typedef LONG_PTR DB_LRESERVE;
 #include "binres.idl"
 #include "crtrow.idl"
 #include "errrec.idl"
-#include "errlup.idl"
 #include "getdts.idl"
 #include "cmdstrm.idl"
 #include "cmdprp.idl"
@@ -87,8 +82,6 @@ typedef LONG_PTR DB_LRESERVE;
 #include "rstchg.idl"
 #include "rstupd.idl"
 #include "rstidn.idl"
-#include "rstind.idl"
-#include "rstcridx.idl"
 #include "mulres.idl"
 #include "transact.idl"
 #include "trnjoi.idl"
@@ -98,5 +91,5 @@ typedef LONG_PTR DB_LRESERVE;
 
 cpp_quote("#include <accctrl.h>")
 
-cpp_quote("#pragma pack(pop)")
+cpp_quote("#include <poppack.h>")
 cpp_quote("")
