diff --git a/include/pdh.h b/include/pdh.h
index 5ef76b07abe4a2a5e7827376705512304bd5f574..32e29acedf0d5b629c4960324ae0cd4871ae30ff 100644
--- a/include/pdh.h
+++ b/include/pdh.h
@@ -189,6 +189,18 @@ typedef struct _PDH_COUNTER_INFO_W
     DWORD DataBuffer[1];
 } PDH_COUNTER_INFO_W, *PPDH_COUNTER_INFO_W;
 
+typedef struct _PDH_FMT_COUNTERVALUE_ITEM_A
+{
+    LPSTR szName;
+    PDH_FMT_COUNTERVALUE FmtValue;
+} PDH_FMT_COUNTERVALUE_ITEM_A, *PPDH_FMT_COUNTERVALUE_ITEM_A;
+
+typedef struct _PDH_FMT_COUNTERVALUE_ITEM_W
+{
+    LPWSTR szName;
+    PDH_FMT_COUNTERVALUE FmtValue;
+} PDH_FMT_COUNTERVALUE_ITEM_W, *PPDH_FMT_COUNTERVALUE_ITEM_W;
+
 typedef struct _PDH_RAW_COUNTER_ITEM_A
 {
     LPSTR szName;
