diff --git a/include/propsys.idl b/include/propsys.idl
index 1cfcaf338da506fbcd2d6f2c5a5f04041961c768..9dc93caf15d65a21e6fcbe9060ef559d887849a8 100644
--- a/include/propsys.idl
+++ b/include/propsys.idl
@@ -807,10 +807,7 @@ cpp_quote("#define PKEYSTR_MAX (GUIDSTRING_MAX + 1 + PKEY_PIDSTR_MAX)")
 
 cpp_quote("PSSTDAPI PSCreateMemoryPropertyStore(REFIID,void **);")
 cpp_quote("PSSTDAPI PSCreatePropertyStoreFromObject(IUnknown*,DWORD,REFIID,void **);")
-cpp_quote("PSSTDAPI PSGetPropertySystem(REFIID, void**);")
 cpp_quote("PSSTDAPI PSStringFromPropertyKey(REFPROPERTYKEY,LPWSTR,UINT);")
-cpp_quote("PSSTDAPI PSGetPropertyKeyFromName(PCWSTR,PROPERTYKEY *);")
-cpp_quote("PSSTDAPI PSGetNameFromPropertyKey(REFPROPERTYKEY,LPWSTR *);")
 cpp_quote("PSSTDAPI PSPropertyKeyFromString(LPCWSTR,PROPERTYKEY*);")
 cpp_quote("PSSTDAPI PSGetPropertyDescription(REFPROPERTYKEY,REFIID,void **);")
 cpp_quote("PSSTDAPI PSGetPropertyDescriptionListFromString(LPCWSTR,REFIID,void **);")
