diff --git a/include/propvarutil.h b/include/propvarutil.h
index 952f0effd37c2d4b5ab1694c6642ddac53de1fe7..0c23c7caa70e7985608a4e1d88b5fc780a6bb1c5 100644
--- a/include/propvarutil.h
+++ b/include/propvarutil.h
@@ -78,14 +78,6 @@ enum tagPROPVAR_COMPARE_FLAGS
 
 typedef int PROPVAR_COMPARE_FLAGS;
 
-enum tagPSTIME_FLAGS
-{
-    PSTF_UTC   = 0x00000000,
-    PSTF_LOCAL = 0x00000001,
-};
-
-typedef int PSTIME_FLAGS;
-
 PSSTDAPI PropVariantChangeType(PROPVARIANT *ppropvarDest, REFPROPVARIANT propvarSrc,
                                      PROPVAR_CHANGE_FLAGS flags, VARTYPE vt);
 PSSTDAPI InitPropVariantFromGUIDAsString(REFGUID guid, PROPVARIANT *ppropvar);
@@ -94,6 +86,7 @@ PSSTDAPI InitVariantFromGUIDAsString(REFGUID guid, VARIANT *pvar);
 PSSTDAPI InitPropVariantFromBuffer(const VOID *pv, UINT cb, PROPVARIANT *ppropvar);
 PSSTDAPI InitPropVariantFromCLSID(REFCLSID clsid, PROPVARIANT *ppropvar);
 PSSTDAPI InitVariantFromBuffer(const VOID *pv, UINT cb, VARIANT *pvar);
+PSSTDAPI PropVariantToGUID(const PROPVARIANT *ppropvar, GUID *guid);
 PSSTDAPI VariantToGUID(const VARIANT *pvar, GUID *guid);
 PSSTDAPI_(INT) PropVariantCompareEx(REFPROPVARIANT propvar1, REFPROPVARIANT propvar2,
                                 PROPVAR_COMPARE_UNIT uint, PROPVAR_COMPARE_FLAGS flags);
@@ -101,14 +94,11 @@ PSSTDAPI InitPropVariantFromFileTime(const FILETIME *pftIn, PROPVARIANT *ppropva
 PSSTDAPI InitPropVariantFromStringVector(PCWSTR *strs, ULONG count, PROPVARIANT *ppropvar);
 
 PSSTDAPI PropVariantToDouble(REFPROPVARIANT propvarIn, double *ret);
-PSSTDAPI PropVariantToFileTime(REFPROPVARIANT propvar, PSTIME_FLAGS flags, FILETIME *out);
-PSSTDAPI PropVariantToGUID(const PROPVARIANT *ppropvar, GUID *guid);
 PSSTDAPI PropVariantToInt16(REFPROPVARIANT propvarIn, SHORT *ret);
 PSSTDAPI PropVariantToInt32(REFPROPVARIANT propvarIn, LONG *ret);
 PSSTDAPI PropVariantToInt64(REFPROPVARIANT propvarIn, LONGLONG *ret);
 PSSTDAPI PropVariantToUInt16(REFPROPVARIANT propvarIn, USHORT *ret);
 PSSTDAPI PropVariantToUInt32(REFPROPVARIANT propvarIn, ULONG *ret);
-PSSTDAPI PropVariantToUInt32Vector(REFPROPVARIANT propvar, ULONG *buffer, ULONG size, ULONG *count);
 PSSTDAPI_(ULONG) PropVariantToUInt32WithDefault(REFPROPVARIANT propvarIn, ULONG uLDefault);
 PSSTDAPI PropVariantToUInt64(REFPROPVARIANT propvarIn, ULONGLONG *ret);
 PSSTDAPI PropVariantToBoolean(REFPROPVARIANT propvarIn, BOOL *ret);
@@ -124,8 +114,6 @@ PSSTDAPI PropVariantToStringAlloc(REFPROPVARIANT propvarIn, WCHAR **ret);
 PSSTDAPI PropVariantToVariant(const PROPVARIANT *propvar, VARIANT *var);
 PSSTDAPI VariantToPropVariant(const VARIANT* var, PROPVARIANT* propvar);
 
-PSSTDAPI PropVariantGetStringElem(const PROPVARIANT *propvar, ULONG idx, WCHAR **ret);
-
 #ifdef __cplusplus
 
 HRESULT InitPropVariantFromBoolean(BOOL fVal, PROPVARIANT *ppropvar);
