diff --git a/include/pshpack1.h b/include/pshpack1.h
index 573ca0e4cdd4b61c9969c024eac6c4e91d5a14e3..f9d4caa149a8e3ed00317fd3772317e3f7f58cde 100644
--- a/include/pshpack1.h
+++ b/include/pshpack1.h
@@ -16,8 +16,92 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#ifdef _MSC_VER
-# pragma warning(disable:4103)
-#endif
+#if defined(__WINE_PSHPACK_H15)
+
+   /* Depth > 15 */
+#  error "Alignment nesting > 15 is not supported"
+
+#else
+
+#  if !defined(__WINE_PSHPACK_H)
+#    define __WINE_PSHPACK_H  1
+     /* Depth == 1 */
+#  elif !defined(__WINE_PSHPACK_H2)
+#    define __WINE_PSHPACK_H2 1
+     /* Depth == 2 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H3)
+#    define __WINE_PSHPACK_H3 1
+     /* Depth == 3 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H4)
+#    define __WINE_PSHPACK_H4 1
+     /* Depth == 4 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H5)
+#    define __WINE_PSHPACK_H5 1
+     /* Depth == 5 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H6)
+#    define __WINE_PSHPACK_H6 1
+     /* Depth == 6 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H7)
+#    define __WINE_PSHPACK_H7 1
+     /* Depth == 7 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H8)
+#    define __WINE_PSHPACK_H8 1
+     /* Depth == 8 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H9)
+#    define __WINE_PSHPACK_H9 1
+     /* Depth == 9 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H10)
+#    define __WINE_PSHPACK_H10 1
+     /* Depth == 10 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H11)
+#    define __WINE_PSHPACK_H11 1
+     /* Depth == 11 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H12)
+#    define __WINE_PSHPACK_H12 1
+     /* Depth == 12 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H13)
+#    define __WINE_PSHPACK_H13 1
+     /* Depth == 13 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H14)
+#    define __WINE_PSHPACK_H14 1
+     /* Depth == 14 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  elif !defined(__WINE_PSHPACK_H15)
+#    define __WINE_PSHPACK_H15 1
+     /* Depth == 15 */
+#    define __WINE_INTERNAL_POPPACK
+#    include <poppack.h>
+#  endif
 
-#pragma pack(push,1)
+#  ifdef _MSC_VER
+#   pragma warning(disable:4103)
+#  endif
+
+#  pragma pack(1)
+
+#endif
