diff --git a/include/ras.h b/include/ras.h
index b3d4f6f235a301c92921bd9850f7f25c5b953ca4..210a2a45f09828dd948343eccded7f4c60feb1ad 100644
--- a/include/ras.h
+++ b/include/ras.h
@@ -24,7 +24,7 @@
 #ifdef __cplusplus
 extern "C" {
 #endif
-#pragma pack(push,4)
+#include <pshpack4.h>
 #include <inaddr.h>
 #include <in6addr.h>
 
@@ -551,7 +551,7 @@ DWORD WINAPI RasSetAutodialEnableA(DWORD dwDialingLocation, BOOL fEnabled);
 DWORD WINAPI RasSetAutodialEnableW(DWORD dwDialingLocation, BOOL fEnabled);
 #define RasSetAutodialEnable WINELIB_NAME_AW(RasSetAutodialEnable)
 
-#pragma pack(pop)
+#include <poppack.h>
 #ifdef __cplusplus
 }
 #endif
