diff --git a/include/richedit.h b/include/richedit.h
index 6c073539d883d8c950fdef597f8c5e1f28ac42f5..2709d787502c2541a582c078aea8be252247ad7a 100644
--- a/include/richedit.h
+++ b/include/richedit.h
@@ -20,7 +20,7 @@
 #ifndef __WINE_RICHEDIT_H
 #define __WINE_RICHEDIT_H
 
-#pragma pack(push,4)
+#include <pshpack4.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -1083,6 +1083,6 @@ typedef struct _settextex {
 }
 #endif
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* __WINE_RICHEDIT_H */
