diff --git a/include/roerrorapi.h b/include/roerrorapi.h
index 8f3200c559da77e78116b4ac99bf5fe893edf100..741347455b174e1d7dd9ff8ef1d89be6cd110944 100644
--- a/include/roerrorapi.h
+++ b/include/roerrorapi.h
@@ -33,12 +33,8 @@ typedef enum
 } RO_ERROR_REPORTING_FLAGS;
 
 HRESULT WINAPI GetRestrictedErrorInfo(IRestrictedErrorInfo **info);
-HRESULT WINAPI RoGetErrorReportingFlags(UINT32 *flags);
 BOOL    WINAPI RoOriginateError(HRESULT error, HSTRING message);
-BOOL    WINAPI RoOriginateErrorW(HRESULT error, UINT max_len, const WCHAR *message);
 BOOL    WINAPI RoOriginateLanguageException(HRESULT error, HSTRING message, IUnknown *language_exception);
-HRESULT WINAPI RoReportUnhandledError(IRestrictedErrorInfo *info);
 HRESULT WINAPI RoSetErrorReportingFlags(UINT32 flags);
-HRESULT WINAPI SetRestrictedErrorInfo(IRestrictedErrorInfo *info);
 
 #endif /* _ROERROR_H */
