diff --git a/include/rpc.h b/include/rpc.h
index 88bd51e94bf2f156514e93ce223f2e16dad45665..2d4323428c47679241c80723b4d0b15988f1eee6 100644
--- a/include/rpc.h
+++ b/include/rpc.h
@@ -79,7 +79,7 @@ typedef void* I_RPC_HANDLE;
 #define RpcTryFinally
 #define RpcFinally
 #define RpcEndFinally
-#define RpcExceptionCode() ((ULONG)0)
+#define RpcExceptionCode() 0
 /* #define RpcAbnormalTermination() abort() */
 
 #endif /* USE_COMPILER_EXCEPTIONS */
