diff --git a/include/shlwapi.h b/include/shlwapi.h
index e127d0ac578ce04337d7c8f5ff99de6ad14ff9c7..c643d2a4e3a73164431202fc43e1ae7bd1e4488f 100644
--- a/include/shlwapi.h
+++ b/include/shlwapi.h
@@ -29,7 +29,7 @@
 extern "C" {
 #endif /* defined(__cplusplus) */
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 #ifndef WINSHLWAPI
 #ifndef _SHLWAPI_
@@ -406,10 +406,6 @@ WINSHLWAPI BOOL WINAPI PathFileExistsAndAttributesA(LPCSTR,DWORD*);
 WINSHLWAPI BOOL WINAPI PathFileExistsAndAttributesW(LPCWSTR,DWORD*);
 #define PathFileExistsAndAttributes WINELIB_NAME_AW(PathFileExistsAndAttributes)
 
-WINSHLWAPI BOOL WINAPI PathFileExistsDefExtA(LPSTR,DWORD);
-WINSHLWAPI BOOL WINAPI PathFileExistsDefExtW(LPWSTR,DWORD);
-#define PathFileExistsDefExt WINELIB_NAME_AW(PathFileExistsDefExt)
-
 WINSHLWAPI LPSTR  WINAPI PathFindExtensionA(LPCSTR);
 WINSHLWAPI LPWSTR WINAPI PathFindExtensionW(LPCWSTR);
 #define PathFindExtension WINELIB_NAME_AW(PathFindExtension)
@@ -426,10 +422,6 @@ WINSHLWAPI BOOL WINAPI PathFindOnPathA(LPSTR,LPCSTR*);
 WINSHLWAPI BOOL WINAPI PathFindOnPathW(LPWSTR,LPCWSTR*);
 #define PathFindOnPath WINELIB_NAME_AW(PathFindOnPath)
 
-WINSHLWAPI BOOL WINAPI PathFindOnPathExA(LPSTR,LPCSTR*,DWORD);
-WINSHLWAPI BOOL WINAPI PathFindOnPathExW(LPWSTR,LPCWSTR*,DWORD);
-#define PathFindOnPathEx WINELIB_NAME_AW(PathFindOnPathEx)
-
 WINSHLWAPI LPSTR  WINAPI PathGetArgsA(LPCSTR);
 WINSHLWAPI LPWSTR WINAPI PathGetArgsW(LPCWSTR);
 #define PathGetArgs WINELIB_NAME_AW(PathGetArgs)
@@ -1103,7 +1095,7 @@ HRESULT WINAPI DllGetVersion(DLLVERSIONINFO *);
 typedef struct _DLLVERSIONINFO2 {
     DLLVERSIONINFO info1;
     DWORD          dwFlags;    /* Reserved */
-    ULONGLONG      ullVersion; /* 16 bits each for Major, Minor, Build, QFE */
+    ULONGLONG DECLSPEC_ALIGN(8) ullVersion; /* 16 bits each for Major, Minor, Build, QFE */
 } DLLVERSIONINFO2;
 
 #define DLLVER_MAJOR_MASK 0xFFFF000000000000
@@ -1210,7 +1202,7 @@ WINSHLWAPI HRESULT WINAPI SHGetViewStatePropertyBag(PCIDLIST_ABSOLUTE pidl, PCWS
 
 WINSHLWAPI BOOL WINAPI SHIsLowMemoryMachine(DWORD type);
 
-#pragma pack(pop)
+#include <poppack.h>
 
 WINSHLWAPI HANDLE WINAPI SHAllocShared(const void *data, DWORD size, DWORD pid);
 WINSHLWAPI BOOL WINAPI SHFreeShared(HANDLE handle, DWORD pid);
