diff --git a/include/snmp.h b/include/snmp.h
index 2d9853bb7ea19d948f9390d6448fa20b257164d7..026c6f0367fb1693cae0d14d19a56db472ab0e73 100644
--- a/include/snmp.h
+++ b/include/snmp.h
@@ -20,7 +20,7 @@
 
 #include <windows.h>
 
-#pragma pack(push,4)
+#include <pshpack4.h>
 
 typedef struct {
     BYTE *stream;
@@ -78,7 +78,7 @@ typedef struct {
     UINT         len;
 } SnmpVarBindList;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #define ASN_UNIVERSAL   0x00
 #define ASN_APPLICATION 0x40
