diff --git a/include/srrestoreptapi.h b/include/srrestoreptapi.h
index 0a2a5b3d4092aeae7d38176f3f01d9e3f80c8ecf..45a027c6be087fbef1a680f9625d4a8783b1db59 100644
--- a/include/srrestoreptapi.h
+++ b/include/srrestoreptapi.h
@@ -31,7 +31,7 @@
 #define MAX_DESC                    64
 #define MAX_DESC_W                  256
 
-#pragma pack(push,1)
+#pragma pack(1)
 
 typedef struct _RESTOREPTINFOA {
     DWORD dwEventType;
@@ -60,8 +60,6 @@ typedef struct _SMGRSTATUS {
     INT64 llSequenceNumber;
 } STATEMGRSTATUS, *PSTATEMGRSTATUS;
 
-#pragma pack(pop)
-
 #ifdef __cplusplus
 extern "C" {
 #endif
