diff --git a/include/unknwn.idl b/include/unknwn.idl
index 7c4a8f0201552d57412658a98a102de6852db228..0798ccf79ab60f7f4475d6edccb7a774858a235f 100644
--- a/include/unknwn.idl
+++ b/include/unknwn.idl
@@ -60,9 +60,6 @@ cpp_quote("        }")
 cpp_quote("        END_INTERFACE")
 cpp_quote("    };")
 cpp_quote("}")
-cpp_quote("#ifdef __CRT_UUID_DECL")
-cpp_quote("__CRT_UUID_DECL(IUnknown, 0x00000000, 0x0000, 0x0000, 0xc0,0x00, 0x00,0x00,0x00,0x00,0x00,0x46)")
-cpp_quote("#endif")
 cpp_quote("#else")
   HRESULT QueryInterface(
     [in] REFIID riid,
