diff --git a/include/wincodec.idl b/include/wincodec.idl
index 5ff13713867fc7e14e17294f4a11d8c7ea23cc41..6752341c02a9c589212b2955c0ec40e5530c670b 100644
--- a/include/wincodec.idl
+++ b/include/wincodec.idl
@@ -1235,6 +1235,17 @@ interface IWICDdsFrameDecode : IUnknown
         [out, size_is(bufferSize)] BYTE *buffer);
 };
 
+[
+    object,
+    uuid(b9bd430d-28a8-41d3-a1f5-f36ee02840bf)
+]
+interface IWICWineDecoder : IUnknown
+{
+    HRESULT Initialize(
+        [in] IStream *stream,
+        [in] WICDecodeOptions options);
+};
+
 cpp_quote("HRESULT WINAPI WICConvertBitmapSource(REFWICPixelFormatGUID dstFormat, IWICBitmapSource *pISrc, IWICBitmapSource **ppIDst);")
 cpp_quote("HRESULT WINAPI WICCreateBitmapFromSection(UINT width, UINT height, REFWICPixelFormatGUID format, HANDLE section, UINT stride, UINT offset, IWICBitmap **bitmap);")
 cpp_quote("HRESULT WINAPI WICCreateBitmapFromSectionEx(UINT width, UINT height, REFWICPixelFormatGUID format, HANDLE section, UINT stride, UINT offset, WICSectionAccessLevel access, IWICBitmap **bitmap);")
