diff --git a/include/windows.applicationmodel.activation.idl b/include/windows.applicationmodel.activation.idl
index 2ea83d82b1d75dc960c062866ca02b767fe56f7a..3392a5bb316c4e1c89a8bc2c1c915e514c944148 100644
--- a/include/windows.applicationmodel.activation.idl
+++ b/include/windows.applicationmodel.activation.idl
@@ -20,6 +20,7 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
@@ -29,6 +30,7 @@ import "windows.applicationmodel.background.idl";
 import "windows.applicationmodel.core.idl";
 import "windows.storage.idl";
 import "windows.system.idl";
+#endif
 
 namespace Windows.ApplicationModel.Background {
     interface IBackgroundTaskInstance;
