diff --git a/include/windows.applicationmodel.datatransfer.dragdrop.core.idl b/include/windows.applicationmodel.datatransfer.dragdrop.core.idl
index afc9765c5784f411acc61197837bd4544916bd37..6f7aa31413041b2f8a48a85ed9d70d5f6a293dc0 100644
--- a/include/windows.applicationmodel.datatransfer.dragdrop.core.idl
+++ b/include/windows.applicationmodel.datatransfer.dragdrop.core.idl
@@ -20,6 +20,7 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
@@ -28,6 +29,7 @@ import "windows.foundation.idl";
 import "windows.applicationmodel.datatransfer.idl";
 import "windows.applicationmodel.datatransfer.dragdrop.idl";
 import "windows.graphics.imaging.idl";
+#endif
 
 namespace Windows.ApplicationModel.DataTransfer.DragDrop.Core {
     interface ICoreDragDropManager;
