diff --git a/include/windows.applicationmodel.datatransfer.idl b/include/windows.applicationmodel.datatransfer.idl
index 7f84513934ab7104445c734b0b05d00dd6e153b3..7ba383f28d1f98fef94bd41b93196ad26531acfc 100644
--- a/include/windows.applicationmodel.datatransfer.idl
+++ b/include/windows.applicationmodel.datatransfer.idl
@@ -20,6 +20,7 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
@@ -29,6 +30,7 @@ import "windows.security.enterprisedata.idl";
 import "windows.storage.idl";
 import "windows.storage.streams.idl";
 import "windows.ui.idl";
+#endif
 
 namespace Windows.ApplicationModel.DataTransfer {
     typedef enum ClipboardHistoryItemsResultStatus ClipboardHistoryItemsResultStatus;
@@ -113,12 +115,8 @@ namespace Windows.ApplicationModel.DataTransfer {
     runtimeclass TargetApplicationChosenEventArgs;
 
     declare {
-        interface Windows.Foundation.Collections.IKeyValuePair<HSTRING, IInspectable *>;
-        interface Windows.Foundation.Collections.IMap<HSTRING, IInspectable *>;
-        interface Windows.Foundation.Collections.IMapView<HSTRING, IInspectable *>;
         interface Windows.Foundation.Collections.IIterable<Windows.ApplicationModel.DataTransfer.ClipboardHistoryItem *>;
         interface Windows.Foundation.Collections.IIterable<Windows.ApplicationModel.DataTransfer.ShareProvider *>;
-        interface Windows.Foundation.Collections.IIterable<Windows.Foundation.Collections.IKeyValuePair<HSTRING, IInspectable *> *>;
         interface Windows.Foundation.Collections.IIterator<Windows.ApplicationModel.DataTransfer.ClipboardHistoryItem *>;
         interface Windows.Foundation.Collections.IIterator<Windows.ApplicationModel.DataTransfer.ShareProvider *>;
         interface Windows.Foundation.Collections.IVectorView<Windows.ApplicationModel.DataTransfer.ClipboardHistoryItem *>;
