diff --git a/include/windows.devices.input.idl b/include/windows.devices.input.idl
index db390e5fff36ab880c9dba87b15463332ce354d4..492d43c580798a6b5f9bef48419a677b04fb290d 100644
--- a/include/windows.devices.input.idl
+++ b/include/windows.devices.input.idl
@@ -20,11 +20,13 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
 import "windowscontracts.idl";
 import "windows.foundation.idl";
+#endif
 
 namespace Windows.Devices.Input {
     typedef enum PointerDeviceType PointerDeviceType;
