diff --git a/include/windows.foundation.metadata.idl b/include/windows.foundation.metadata.idl
index 9487737e071e07b904ba083b4eee63dd70f78ed4..6488405aa7a4805d40f81b903d763bd087825acd 100644
--- a/include/windows.foundation.metadata.idl
+++ b/include/windows.foundation.metadata.idl
@@ -20,11 +20,13 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
 import "windowscontracts.idl";
 import "windows.foundation.idl";
+#endif
 
 namespace Windows.Foundation.Metadata
 {
