diff --git a/include/windows.gaming.input.custom.idl b/include/windows.gaming.input.custom.idl
index 0811b9910916710944ddb90b75d0786768836529..8347f4663d300891ae18758b97c61ee3c094b53c 100644
--- a/include/windows.gaming.input.custom.idl
+++ b/include/windows.gaming.input.custom.idl
@@ -20,6 +20,7 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
@@ -27,6 +28,7 @@ import "windowscontracts.idl";
 import "windows.foundation.idl";
 import "windows.gaming.input.idl";
 import "windows.storage.streams.idl";
+#endif
 
 namespace Windows.Gaming.Input.Custom {
     typedef enum XusbDeviceSubtype XusbDeviceSubtype;
