diff --git a/include/windows.graphics.holographic.idl b/include/windows.graphics.holographic.idl
index 05731f62b024193311cb88d2c77378b4ee4a7c53..c82ce926e07c07201d8b7f39aff6b04d1b2dc599 100644
--- a/include/windows.graphics.holographic.idl
+++ b/include/windows.graphics.holographic.idl
@@ -303,7 +303,7 @@ namespace Windows.Graphics.Holographic {
     [
         contract(Windows.Foundation.UniversalApiContract, 2.0),
         marshaling_behavior(agile),
-        /* static(Windows.Graphics.Holographic.IHolographicSpaceStatics, Windows.Foundation.UniversalApiContract, 2.0), */
+        static(Windows.Graphics.Holographic.IHolographicSpaceStatics, Windows.Foundation.UniversalApiContract, 2.0),
         static(Windows.Graphics.Holographic.IHolographicSpaceStatics2, Windows.Foundation.UniversalApiContract, 4.0),
         static(Windows.Graphics.Holographic.IHolographicSpaceStatics3, Windows.Foundation.UniversalApiContract, 5.0),
         threading(both)
