diff --git a/include/windows.networking.idl b/include/windows.networking.idl
index 0b341c50af2e2b5ade5542de51eaf8c11b7a34c0..28e0f9831546a42fc238835d86aeb528f12a8866 100644
--- a/include/windows.networking.idl
+++ b/include/windows.networking.idl
@@ -17,16 +17,17 @@
  */
 
 #ifdef __WIDL__
-#pragma makedep header install winmd
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
 import "windowscontracts.idl";
 import "windows.foundation.idl";
 import "windows.networking.connectivity.idl";
+#endif
 
 namespace Windows.Networking.Connectivity
 {
