diff --git a/include/windows.security.credentials.idl b/include/windows.security.credentials.idl
index 35b739f29f35f223a2ca2e5ae4511f7c8e4f825a..ca682baaf0e68553dff2659b35a4b4d620d51638 100644
--- a/include/windows.security.credentials.idl
+++ b/include/windows.security.credentials.idl
@@ -25,51 +25,27 @@ import "asyncinfo.idl";
 import "eventtoken.idl";
 import "windowscontracts.idl";
 import "windows.foundation.idl";
-import "windows.security.cryptography.core.idl";
+/* import "windows.security.cryptography.core.idl"; */
 import "windows.storage.streams.idl";
 import "windows.system.idl";
 
-namespace Windows.Security.Cryptography.Core {
-    typedef enum CryptographicPublicKeyBlobType CryptographicPublicKeyBlobType;
-}
-
 namespace Windows.Security.Credentials {
-    typedef enum KeyCredentialAttestationStatus KeyCredentialAttestationStatus;
     typedef enum KeyCredentialCreationOption KeyCredentialCreationOption;
     typedef enum KeyCredentialStatus KeyCredentialStatus;
 
     interface IKeyCredential;
-    interface IKeyCredentialAttestationResult;
     interface IKeyCredentialManagerStatics;
-    interface IKeyCredentialOperationResult;
     interface IKeyCredentialRetrievalResult;
 
     runtimeclass KeyCredential;
-    runtimeclass KeyCredentialAttestationResult;
     runtimeclass KeyCredentialManager;
-    runtimeclass KeyCredentialOperationResult;
     runtimeclass KeyCredentialRetrievalResult;
 
     declare {
-        interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Security.Credentials.KeyCredentialAttestationResult *>;
-        interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Security.Credentials.KeyCredentialOperationResult *>;
         interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Security.Credentials.KeyCredentialRetrievalResult *>;
-        interface Windows.Foundation.IAsyncOperation<Windows.Security.Credentials.KeyCredentialAttestationResult *>;
-        interface Windows.Foundation.IAsyncOperation<Windows.Security.Credentials.KeyCredentialOperationResult *>;
         interface Windows.Foundation.IAsyncOperation<Windows.Security.Credentials.KeyCredentialRetrievalResult *>;
     }
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0)
-    ]
-    enum KeyCredentialAttestationStatus
-    {
-        Success          = 0,
-        UnknownError     = 1,
-        NotSupported     = 2,
-        TemporaryFailure = 3,
-    };
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0)
     ]
@@ -93,32 +69,6 @@ namespace Windows.Security.Credentials {
         SecurityDeviceLocked    = 6,
     };
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
-        exclusiveto(Windows.Security.Credentials.KeyCredential),
-        uuid(9585ef8d-457b-4847-b11a-fa960bbdb138)
-    ]
-    interface IKeyCredential : IInspectable
-    {
-        [propget] HRESULT Name([out, retval] HSTRING *value);
-        [overload("RetrievePublicKey")] HRESULT RetrievePublicKeyWithDefaultBlobType([out, retval] Windows.Storage.Streams.IBuffer **value);
-        [overload("RetrievePublicKey")] HRESULT RetrievePublicKeyWithBlobType([in] Windows.Security.Cryptography.Core.CryptographicPublicKeyBlobType blobType, [out, retval] Windows.Storage.Streams.IBuffer **value);
-        HRESULT RequestSignAsync([in] Windows.Storage.Streams.IBuffer *data, [out, retval] Windows.Foundation.IAsyncOperation<Windows.Security.Credentials.KeyCredentialOperationResult *> **value);
-        HRESULT GetAttestationAsync([out, retval] Windows.Foundation.IAsyncOperation<Windows.Security.Credentials.KeyCredentialAttestationResult *> **value);
-    }
-
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
-        exclusiveto(Windows.Security.Credentials.KeyCredentialAttestationResult),
-        uuid(78aab3a1-a3c1-4103-b6cc-472c44171cbb)
-    ]
-    interface IKeyCredentialAttestationResult : IInspectable
-    {
-        [propget] HRESULT CertificateChainBuffer([out, retval] Windows.Storage.Streams.IBuffer **value);
-        [propget] HRESULT AttestationBuffer([out, retval] Windows.Storage.Streams.IBuffer **value);
-        [propget] HRESULT Status([out, retval] Windows.Security.Credentials.KeyCredentialAttestationStatus *value);
-    }
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         exclusiveto(Windows.Security.Credentials.KeyCredentialManager),
@@ -134,17 +84,6 @@ namespace Windows.Security.Credentials {
         HRESULT DeleteAsync(HSTRING name, [out, retval] Windows.Foundation.IAsyncAction **operation);
     }
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
-        exclusiveto(Windows.Security.Credentials.KeyCredentialOperationResult),
-        uuid(f53786c1-5261-4cdd-976d-cc909ac71620)
-    ]
-    interface IKeyCredentialOperationResult : IInspectable
-    {
-        [propget] HRESULT Result([out, retval] Windows.Storage.Streams.IBuffer **value);
-        [propget] HRESULT Status([out, retval] Windows.Security.Credentials.KeyCredentialStatus *value);
-    }
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         exclusiveto(Windows.Security.Credentials.KeyCredentialRetrievalResult),
@@ -165,15 +104,6 @@ namespace Windows.Security.Credentials {
         [default] interface Windows.Security.Credentials.IKeyCredential;
     }
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
-        marshaling_behavior(agile)
-    ]
-    runtimeclass KeyCredentialAttestationResult
-    {
-        [default] interface Windows.Security.Credentials.IKeyCredentialAttestationResult;
-    }
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         marshaling_behavior(agile),
@@ -184,15 +114,6 @@ namespace Windows.Security.Credentials {
     {
     }
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
-        marshaling_behavior(agile)
-    ]
-    runtimeclass KeyCredentialOperationResult
-    {
-        [default] interface Windows.Security.Credentials.IKeyCredentialOperationResult;
-    }
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         marshaling_behavior(agile)
