diff --git a/include/windows.storage.fileproperties.idl b/include/windows.storage.fileproperties.idl
index 77c17298d9745e1ad6688e2c80575fb6c027c639..9426d02fefb4fa8914e6d13d3fec4b13bcf96ccc 100644
--- a/include/windows.storage.fileproperties.idl
+++ b/include/windows.storage.fileproperties.idl
@@ -57,6 +57,6 @@ namespace Windows.Storage.FileProperties {
     runtimeclass BasicProperties
     {
         [default] interface Windows.Storage.FileProperties.IBasicProperties;
-        /* interface Windows.Storage.FileProperties.IStorageItemExtraProperties; */
+        interface Windows.Storage.FileProperties.IStorageItemExtraProperties;
     }
 }
