diff --git a/include/windows.storage.idl b/include/windows.storage.idl
index 6f4aa6d803b5123e0f625c077ff40b257c988dad..483fc5a957e34e358672862adf2c421cc2b66f67 100644
--- a/include/windows.storage.idl
+++ b/include/windows.storage.idl
@@ -17,11 +17,9 @@
  */
 
 #ifdef __WIDL__
-#pragma makedep header install winmd
 #pragma winrt ns_prefix
 #endif
 
-#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
@@ -32,11 +30,6 @@ import "windows.storage.fileproperties.idl";
 import "windows.storage.search.idl";
 import "windows.storage.streams.idl";
 import "windows.system.idl";
-#endif
-
-namespace Windows.System {
-    runtimeclass User;
-}
 
 namespace Windows.Storage {
     typedef enum ApplicationDataCreateDisposition ApplicationDataCreateDisposition;
@@ -65,11 +58,11 @@ namespace Windows.Storage {
     interface IKnownFoldersStatics4;
     interface ISetVersionDeferral;
     interface ISetVersionRequest;
-    interface IStorageFileStatics;
-    interface IStorageFileStatics2;
     interface IStorageFolder;
     interface IStorageFolderStatics;
     interface IStorageFolderStatics2;
+    interface IStorageFileStatics;
+    interface IStorageFileStatics2;
     interface IStorageItem;
 
     runtimeclass ApplicationData;
@@ -84,20 +77,17 @@ namespace Windows.Storage {
     declare {
         interface Windows.Foundation.Collections.IKeyValuePair<HSTRING, Windows.Storage.ApplicationDataContainer *>;
         interface Windows.Foundation.Collections.IIterable<Windows.Foundation.Collections.IKeyValuePair<HSTRING, Windows.Storage.ApplicationDataContainer *> *>;
-        interface Windows.Foundation.Collections.IIterator<Windows.Foundation.Collections.IKeyValuePair<HSTRING, Windows.Storage.ApplicationDataContainer *> *>;
-        interface Windows.Foundation.Collections.IMapView<HSTRING, Windows.Storage.ApplicationDataContainer *>;
-        interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Storage.ApplicationData *>;
-        interface Windows.Foundation.IAsyncOperation<Windows.Storage.ApplicationData *>;
-        interface Windows.Foundation.TypedEventHandler<Windows.Storage.ApplicationData *, IInspectable *>;
-#ifndef _WINDOWS_APPLICATIONDATA
         interface Windows.Foundation.Collections.IIterable<Windows.Storage.IStorageItem *>;
+        interface Windows.Foundation.Collections.IIterator<Windows.Foundation.Collections.IKeyValuePair<HSTRING, Windows.Storage.ApplicationDataContainer *> *>;
         interface Windows.Foundation.Collections.IIterator<Windows.Storage.IStorageItem *>;
+        interface Windows.Foundation.Collections.IMapView<HSTRING, Windows.Storage.ApplicationDataContainer *>;
         interface Windows.Foundation.Collections.IVectorView<Windows.Storage.IStorageItem *>;
         interface Windows.Foundation.Collections.IVectorView<Windows.Storage.StorageFile *>;
         interface Windows.Foundation.Collections.IVectorView<Windows.Storage.StorageFolder *>;
         interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Foundation.Collections.IVectorView<Windows.Storage.IStorageItem *> *>;
         interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Foundation.Collections.IVectorView<Windows.Storage.StorageFile *> *>;
         interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Foundation.Collections.IVectorView<Windows.Storage.StorageFolder *> *>;
+        interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Storage.ApplicationData *>;
         interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Storage.KnownFoldersAccessStatus>;
         interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Storage.IStorageItem *>;
         interface Windows.Foundation.AsyncOperationCompletedHandler<Windows.Storage.StorageFile *>;
@@ -106,12 +96,13 @@ namespace Windows.Storage {
         interface Windows.Foundation.IAsyncOperation<Windows.Foundation.Collections.IVectorView<Windows.Storage.IStorageItem *> *>;
         interface Windows.Foundation.IAsyncOperation<Windows.Foundation.Collections.IVectorView<Windows.Storage.StorageFile *> *>;
         interface Windows.Foundation.IAsyncOperation<Windows.Foundation.Collections.IVectorView<Windows.Storage.StorageFolder *> *>;
+        interface Windows.Foundation.IAsyncOperation<Windows.Storage.ApplicationData *>;
         interface Windows.Foundation.IAsyncOperation<Windows.Storage.KnownFoldersAccessStatus>;
         interface Windows.Foundation.IAsyncOperation<Windows.Storage.IStorageItem *>;
         interface Windows.Foundation.IAsyncOperation<Windows.Storage.StorageFile *>;
         interface Windows.Foundation.IAsyncOperation<Windows.Storage.StorageFolder *>;
         interface Windows.Foundation.IAsyncOperation<Windows.Storage.StorageStreamTransaction *>;
-#endif
+        interface Windows.Foundation.TypedEventHandler<Windows.Storage.ApplicationData *, IInspectable *>;
     }
 
     [
@@ -136,7 +127,6 @@ namespace Windows.Storage {
         SharedLocal = 4,
     };
 
-#ifndef _WINDOWS_APPLICATIONDATA
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0)
     ]
@@ -242,7 +232,6 @@ namespace Windows.Storage {
         File   = 0x1,
         Folder = 0x2,
     };
-#endif
 
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
@@ -356,7 +345,6 @@ namespace Windows.Storage {
         );
     }
 
-#ifndef _WINDOWS_APPLICATIONDATA
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         exclusiveto(Windows.Storage.KnownFolders),
@@ -450,7 +438,6 @@ namespace Windows.Storage {
             [out, retval] Windows.Foundation.IAsyncOperation<Windows.Storage.StorageFolder *> **operation
         );
     }
-#endif
 
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
@@ -474,7 +461,6 @@ namespace Windows.Storage {
         HRESULT GetDeferral([out, retval] Windows.Storage.SetVersionDeferral **deferral);
     }
 
-#ifndef _WINDOWS_APPLICATIONDATA
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         uuid(fa3f6186-4214-428c-a64c-14c9ac7315ea)
@@ -605,7 +591,6 @@ namespace Windows.Storage {
         [propget] HRESULT Stream([out, retval] Windows.Storage.Streams.IRandomAccessStream **value);
         HRESULT CommitAsync([out, retval] Windows.Foundation.IAsyncAction **operation);
     }
-#endif
 
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
@@ -632,7 +617,6 @@ namespace Windows.Storage {
         [contract(Windows.Foundation.UniversalApiContract, 12.0)] interface Windows.Foundation.IClosable;
     }
 
-#ifndef _WINDOWS_APPLICATIONDATA
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         marshaling_behavior(agile),
@@ -668,8 +652,8 @@ namespace Windows.Storage {
 
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
-        /* static(Windows.Storage.IStorageFileStatics, Windows.Foundation.UniversalApiContract, 1.0), */
-        /* static(Windows.Storage.IStorageFileStatics2, Windows.Foundation.UniversalApiContract, 10.0) */
+        static(Windows.Storage.IStorageFileStatics, Windows.Foundation.UniversalApiContract, 1.0),
+        static(Windows.Storage.IStorageFileStatics2, Windows.Foundation.UniversalApiContract, 10.0)
     ]
     runtimeclass StorageFile
     {
@@ -677,30 +661,30 @@ namespace Windows.Storage {
         interface Windows.Storage.Streams.IInputStreamReference;
         interface Windows.Storage.Streams.IRandomAccessStreamReference;
         interface Windows.Storage.IStorageItem;
-        /* interface Windows.Storage.IStorageItemProperties; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItemProperties2; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItem2; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItemPropertiesWithProvider; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageFilePropertiesWithAvailability; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageFile2; */
+        interface Windows.Storage.IStorageItemProperties;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItemProperties2;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItem2;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItemPropertiesWithProvider;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageFilePropertiesWithAvailability;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageFile2;
     }
 
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
-        /* static(Windows.Storage.IStorageFolderStatics, Windows.Foundation.UniversalApiContract, 1.0), */
-        /* static(Windows.Storage.IStorageFolderStatics2, Windows.Foundation.UniversalApiContract, 10.0) */
+        static(Windows.Storage.IStorageFolderStatics, Windows.Foundation.UniversalApiContract, 1.0),
+        static(Windows.Storage.IStorageFolderStatics2, Windows.Foundation.UniversalApiContract, 10.0)
     ]
     runtimeclass StorageFolder
     {
         [default] interface Windows.Storage.IStorageFolder;
         interface Windows.Storage.IStorageItem;
-        /* interface Windows.Storage.Search.IStorageFolderQueryOperations; */
-        /* interface Windows.Storage.IStorageItemProperties; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItemProperties2; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItem2; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageFolder2; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItemPropertiesWithProvider; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 6.0)] interface Windows.Storage.IStorageFolder3; */
+        interface Windows.Storage.Search.IStorageFolderQueryOperations;
+        interface Windows.Storage.IStorageItemProperties;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItemProperties2;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItem2;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageFolder2;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.Storage.IStorageItemPropertiesWithProvider;
+        [contract(Windows.Foundation.UniversalApiContract, 6.0)] interface Windows.Storage.IStorageFolder3;
     }
 
     [
@@ -711,5 +695,4 @@ namespace Windows.Storage {
         [default] interface Windows.Storage.IStorageStreamTransaction;
         interface Windows.Foundation.IClosable;
     }
-#endif
 }
