diff --git a/include/windows.storage.streams.idl b/include/windows.storage.streams.idl
index 161baa88e41a0e61433298815390486f93cc4d9c..52e2a19f3b012b88edc26c571c6b1d3012fe6182 100644
--- a/include/windows.storage.streams.idl
+++ b/include/windows.storage.streams.idl
@@ -20,9 +20,11 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "eventtoken.idl";
 import "windows.foundation.idl";
+#endif
 
 namespace Windows.Storage {
     interface IStorageFile;
@@ -314,6 +316,7 @@ namespace Windows.Storage.Streams {
     {
     }
 
+#ifndef _WINDOWS_STORAGE
     [
         activatable(Windows.Storage.Streams.IBufferFactory, Windows.Foundation.UniversalApiContract, 1.0),
         contract(Windows.Foundation.UniversalApiContract, 1.0),
@@ -347,6 +350,7 @@ namespace Windows.Storage.Streams {
     {
         [default] interface Windows.Foundation.IAsyncOperation<UINT32>;
     }
+#endif
 
     [
         activatable(Windows.Foundation.UniversalApiContract, 1.0),
