diff --git a/include/windows.system.userprofile.idl b/include/windows.system.userprofile.idl
index 5694387af2c0b8629f24ed47856e0f58b4f26265..756302d4887f0047b235105b7edc3a8dbe284b56 100644
--- a/include/windows.system.userprofile.idl
+++ b/include/windows.system.userprofile.idl
@@ -20,12 +20,14 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
 import "windowscontracts.idl";
 import "windows.foundation.idl";
 import "windows.globalization.idl";
+#endif
 
 namespace Windows.System.UserProfile {
 
@@ -62,25 +64,29 @@ namespace Windows.System.UserProfile {
         [propget] HRESULT AdvertisingId([out, retval] HSTRING *value);
     }
 
+#ifndef _TWINAPI_APPCORE
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         marshaling_behavior(agile),
         static(Windows.System.UserProfile.IGlobalizationPreferencesStatics, Windows.Foundation.UniversalApiContract, 1.0),
-        /* static(Windows.System.UserProfile.IGlobalizationPreferencesStatics2, Windows.Foundation.UniversalApiContract, 5.0), */
-        /* static(Windows.System.UserProfile.IGlobalizationPreferencesStatics3, Windows.Foundation.UniversalApiContract, 6.0), */
+        static(Windows.System.UserProfile.IGlobalizationPreferencesStatics2, Windows.Foundation.UniversalApiContract, 5.0),
+        static(Windows.System.UserProfile.IGlobalizationPreferencesStatics3, Windows.Foundation.UniversalApiContract, 6.0),
     ]
     runtimeclass GlobalizationPreferences
     {
     }
+#endif
 
+#ifndef _WINDOWS_GLOBALIZATION
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         marshaling_behavior(standard),
         static(Windows.System.UserProfile.IAdvertisingManagerStatics, Windows.Foundation.UniversalApiContract, 1.0),
-        /* static(Windows.System.UserProfile.IAdvertisingManagerStatics2, Windows.Foundation.UniversalApiContract, 3.0), */
+        static(Windows.System.UserProfile.IAdvertisingManagerStatics2, Windows.Foundation.UniversalApiContract, 3.0),
     ]
     runtimeclass AdvertisingManager
     {
     }
+#endif
 
 }
