diff --git a/include/windows.ui.core.idl b/include/windows.ui.core.idl
index 68159d7c3eeb55384667d5d6e951fa85b03b105c..e6b4ec5007dedf1beedc144c81a48f0e73b486a2 100644
--- a/include/windows.ui.core.idl
+++ b/include/windows.ui.core.idl
@@ -20,6 +20,7 @@
 #pragma winrt ns_prefix
 #endif
 
+#ifndef DO_NO_IMPORTS
 import "inspectable.idl";
 import "asyncinfo.idl";
 import "eventtoken.idl";
@@ -32,6 +33,7 @@ import "windows.ui.input.idl";
 /* import "windows.ui.input.inking.idl"; */
 /* import "windows.ui.input.inking.core.idl"; */
 /* import "windows.ui.popups.idl"; */
+#endif
 
 namespace Windows.UI.Input {
     runtimeclass PointerPoint;
@@ -43,7 +45,6 @@ namespace Windows.UI.Input {
 }
 
 namespace Windows.UI.Core {
-    typedef enum CoreAcceleratorKeyEventType CoreAcceleratorKeyEventType;
     typedef enum CoreCursorType CoreCursorType;
     typedef enum CoreDispatcherPriority CoreDispatcherPriority;
     typedef enum CoreProcessEventsOption CoreProcessEventsOption;
@@ -54,7 +55,6 @@ namespace Windows.UI.Core {
     typedef struct CorePhysicalKeyStatus CorePhysicalKeyStatus;
     typedef struct CoreProximityEvaluation CoreProximityEvaluation;
 
-    interface IAcceleratorKeyEventArgs;
     interface IAutomationProviderRequestedEventArgs;
     interface ICharacterReceivedEventArgs;
     interface IClosestInteractiveBoundsRequestedEventArgs;
@@ -99,7 +99,6 @@ namespace Windows.UI.Core {
     runtimeclass WindowSizeChangedEventArgs;
 
     declare {
-        interface Windows.Foundation.TypedEventHandler<Windows.UI.Core.CoreDispatcher *, Windows.UI.Core.AcceleratorKeyEventArgs *>;
         interface Windows.Foundation.TypedEventHandler<Windows.UI.Core.CoreWindow *, IInspectable *>;
         interface Windows.Foundation.TypedEventHandler<Windows.UI.Core.CoreWindow *, Windows.UI.Core.AutomationProviderRequestedEventArgs *>;
         interface Windows.Foundation.TypedEventHandler<Windows.UI.Core.CoreWindow *, Windows.UI.Core.CharacterReceivedEventArgs *>;
@@ -114,22 +113,6 @@ namespace Windows.UI.Core {
         interface Windows.Foundation.TypedEventHandler<Windows.UI.Core.CoreWindow *, Windows.UI.Core.WindowSizeChangedEventArgs *>;
     }
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0)
-    ]
-    enum CoreAcceleratorKeyEventType
-    {
-        Character           = 2,
-        DeadCharacter       = 3,
-        KeyDown             = 0,
-        KeyUp               = 1,
-        SystemCharacter     = 6,
-        SystemDeadCharacter = 7,
-        SystemKeyDown       = 4,
-        SystemKeyUp         = 5,
-        UnicodeCharacter    = 8,
-    };
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0)
     ]
@@ -253,19 +236,6 @@ namespace Windows.UI.Core {
     ]
     delegate HRESULT IdleDispatchedHandler([in] Windows.UI.Core.IdleDispatchedHandlerArgs *e);
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
-        exclusiveto(Windows.UI.Core.AcceleratorKeyEventArgs),
-        uuid(ff1c4c4a-9287-470b-836e-9086e3126ade)
-    ]
-    interface IAcceleratorKeyEventArgs : IInspectable
-        requires Windows.UI.Core.ICoreWindowEventArgs
-    {
-        [propget] HRESULT EventType([out, retval] Windows.UI.Core.CoreAcceleratorKeyEventType *value);
-        [propget] HRESULT VirtualKey([out, retval] Windows.System.VirtualKey *value);
-        [propget] HRESULT KeyStatus([out, retval] Windows.UI.Core.CorePhysicalKeyStatus *value);
-    }
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         exclusiveto(Windows.UI.Core.AutomationProviderRequestedEventArgs),
@@ -303,16 +273,6 @@ namespace Windows.UI.Core {
         [propput] HRESULT ClosestInteractiveBounds([in] Windows.Foundation.Rect value);
     }
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
-        uuid(9ffdf7f5-b8c9-4ef0-b7d2-1de626561fc8)
-    ]
-    interface ICoreAcceleratorKeys : IInspectable
-    {
-        [eventadd] HRESULT AcceleratorKeyActivated([in] Windows.Foundation.TypedEventHandler<Windows.UI.Core.CoreDispatcher *, Windows.UI.Core.AcceleratorKeyEventArgs *> *handler, [out, retval] EventRegistrationToken *cookie);
-        [eventremove] HRESULT AcceleratorKeyActivated([in] EventRegistrationToken cookie);
-    }
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         exclusiveto(Windows.UI.Core.CoreCursor),
@@ -581,16 +541,6 @@ namespace Windows.UI.Core {
         [propget] HRESULT Size([out, retval] Windows.Foundation.Size *value);
     }
 
-    [
-        contract(Windows.Foundation.UniversalApiContract, 1.0),
-        marshaling_behavior(agile)
-    ]
-    runtimeclass AcceleratorKeyEventArgs
-    {
-        [default] interface Windows.UI.Core.IAcceleratorKeyEventArgs;
-        interface Windows.UI.Core.ICoreWindowEventArgs;
-    }
-
     [
         contract(Windows.Foundation.UniversalApiContract, 1.0),
         marshaling_behavior(standard)
@@ -638,8 +588,8 @@ namespace Windows.UI.Core {
     {
         [default] interface Windows.UI.Core.ICoreDispatcher;
         interface Windows.UI.Core.ICoreAcceleratorKeys;
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.UI.Core.ICoreDispatcherWithTaskPriority; */
-        /* [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.UI.Core.ICoreDispatcher2; */
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.UI.Core.ICoreDispatcherWithTaskPriority;
+        [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.UI.Core.ICoreDispatcher2;
     }
 
     [
@@ -651,11 +601,11 @@ namespace Windows.UI.Core {
     {
         [default] interface Windows.UI.Core.ICoreWindow;
         [contract(Windows.Foundation.UniversalApiContract, 1.0)] interface Windows.UI.Core.ICoreWindow2;
-        /* [contract(Windows.Foundation.UniversalApiContract, 2.0)] interface Windows.UI.Core.ICorePointerRedirector; */
+        [contract(Windows.Foundation.UniversalApiContract, 2.0)] interface Windows.UI.Core.ICorePointerRedirector;
         [contract(Windows.Foundation.UniversalApiContract, 3.0)] interface Windows.UI.Core.ICoreWindow3;
         [contract(Windows.Foundation.UniversalApiContract, 4.0)] interface Windows.UI.Core.ICoreWindow4;
         [contract(Windows.Foundation.UniversalApiContract, 5.0)] interface Windows.UI.Core.ICoreWindow5;
-        /* [contract(Windows.Foundation.UniversalApiContract, 8.0)] interface Windows.UI.Core.ICoreWindowWithContext; */
+        [contract(Windows.Foundation.UniversalApiContract, 8.0)] interface Windows.UI.Core.ICoreWindowWithContext;
     }
 
     [
