diff --git a/include/wine/rbtree.h b/include/wine/rbtree.h
index 81367f3f7c4a6c157285c1de59ccff2c2ceb3cd9..12004010a850d9eb6dc9aa86aa1f48211be5f975 100644
--- a/include/wine/rbtree.h
+++ b/include/wine/rbtree.h
@@ -20,6 +20,10 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
+#if 0
+#pragma makedep install
+#endif
+
 #ifndef __WINE_WINE_RBTREE_H
 #define __WINE_WINE_RBTREE_H
 
