diff --git a/include/wine/server_protocol.h b/include/wine/server_protocol.h
index dc1ceeb142e6358ee83628e1191ec4dc8222380c..a39ab12397076549e2340acfcb6a4e8ac27752a8 100644
--- a/include/wine/server_protocol.h
+++ b/include/wine/server_protocol.h
@@ -18,7 +18,6 @@
 
 typedef unsigned int obj_handle_t;
 typedef unsigned int user_handle_t;
-typedef unsigned int d3dkmt_handle_t;
 typedef unsigned int atom_t;
 typedef unsigned int process_id_t;
 typedef unsigned int thread_id_t;
@@ -128,7 +127,7 @@ struct context_data
     union
     {
         struct { unsigned int eip, ebp, esp, eflags, cs, ss; } i386_regs;
-        struct { unsigned __int64 rip, rsp;
+        struct { unsigned __int64 rip, rbp, rsp;
                  unsigned int cs, ss, flags, __pad; } x86_64_regs;
         struct { unsigned int sp, lr, pc, cpsr; } arm_regs;
         struct { unsigned __int64 sp, pc, pstate; } arm64_regs;
@@ -136,7 +135,7 @@ struct context_data
     union
     {
         struct { unsigned int eax, ebx, ecx, edx, esi, edi; } i386_regs;
-        struct { unsigned __int64 rax, rbx, rcx, rdx, rbp, rsi, rdi,
+        struct { unsigned __int64 rax,rbx, rcx, rdx, rsi, rdi,
                                   r8, r9, r10, r11, r12, r13, r14, r15; } x86_64_regs;
         struct { unsigned int r[13]; } arm_regs;
         struct { unsigned __int64 x[31]; } arm64_regs;
@@ -492,14 +491,11 @@ enum apc_type
 struct user_apc
 {
     enum apc_type    type;
-    unsigned int     flags;
+    int              __pad;
     client_ptr_t     func;
     apc_param_t      args[3];
 };
 
-#define SERVER_USER_APC_SPECIAL               0x01
-#define SERVER_USER_APC_CALLBACK_DATA_CONTEXT 0x02
-
 union apc_call
 {
     enum apc_type type;
@@ -831,7 +827,6 @@ struct pe_image_info
     mem_size_t     stack_commit;
     unsigned int   entry_point;
     unsigned int   map_size;
-    unsigned int   alignment;
     unsigned int   zerobits;
     unsigned int   subsystem;
     unsigned short subsystem_minor;
@@ -849,7 +844,6 @@ struct pe_image_info
     unsigned char  image_flags;
     unsigned int   loader_flags;
     unsigned int   header_size;
-    unsigned int   header_map_size;
     unsigned int   file_size;
     unsigned int   checksum;
     unsigned int   dbg_offset;
@@ -953,30 +947,9 @@ union udp_endpoint
     } ipv6;
 };
 
-enum d3dkmt_type
-{
-    D3DKMT_ADAPTER      = 1,
-    D3DKMT_DEVICE       = 2,
-    D3DKMT_SOURCE       = 3,
-    D3DKMT_MUTEX        = 4,
-    D3DKMT_SYNC         = 5,
-    D3DKMT_RESOURCE     = 6,
-    D3DKMT_ALLOCATION   = 7,
-};
-
 
 
 
-struct obj_locator
-{
-    object_id_t          id;
-    mem_size_t           offset;
-};
-
-#define MAX_ATOM_LEN     255
-#define WH_WINEVENT      (WH_MAXHOOK + 1)
-#define NB_HOOKS         (WH_WINEVENT - WH_MINHOOK + 1)
-
 struct shared_cursor
 {
     int                  x;
@@ -991,18 +964,15 @@ typedef volatile struct
     struct shared_cursor cursor;
     unsigned char        keystate[256];
     unsigned __int64     monitor_serial;
-    unsigned __int64     keystate_serial;
 } desktop_shm_t;
 
 typedef volatile struct
 {
-    timeout_t            access_time;
+    int                  hooks_count[WH_MAX - WH_MIN + 2];
     unsigned int         wake_mask;
     unsigned int         wake_bits;
     unsigned int         changed_mask;
     unsigned int         changed_bits;
-    unsigned int         internal_bits;
-    int                  hooks_count[NB_HOOKS];
 } queue_shm_t;
 
 typedef volatile struct
@@ -1019,36 +989,13 @@ typedef volatile struct
     int                  cursor_count;
     unsigned char        keystate[256];
     int                  keystate_lock;
-    unsigned __int64     keystate_serial;
 } input_shm_t;
 
-typedef volatile struct
-{
-    atom_t               atom;
-    unsigned int         style;
-    unsigned int         cls_extra;
-    unsigned int         win_extra;
-    mod_handle_t         instance;
-    data_size_t          name_offset;
-    data_size_t          name_len;
-    WCHAR                name[MAX_ATOM_LEN];
-    unsigned short       __pad;
-    char                 extra[];
-} class_shm_t;
-
-typedef volatile struct
-{
-    struct obj_locator   class;
-    unsigned int         dpi_context;
-} window_shm_t;
-
 typedef volatile union
 {
     desktop_shm_t        desktop;
     queue_shm_t          queue;
     input_shm_t          input;
-    class_shm_t          class;
-    window_shm_t         window;
 } object_shm_t;
 
 typedef volatile struct
@@ -1058,10 +1005,11 @@ typedef volatile struct
     object_shm_t         shm;
 } shared_object_t;
 
-typedef volatile struct
+struct obj_locator
 {
-    struct user_entry user_entries[MAX_USER_HANDLES];
-} session_shm_t;
+    object_id_t          id;
+    mem_size_t           offset;
+};
 
 
 
@@ -1155,6 +1103,7 @@ struct init_process_done_request
     char __pad_12[4];
     client_ptr_t teb;
     client_ptr_t peb;
+    client_ptr_t ldt_copy;
 };
 struct init_process_done_reply
 {
@@ -1181,10 +1130,8 @@ struct init_first_thread_reply
     thread_id_t  tid;
     timeout_t    server_start;
     unsigned int session_id;
-    obj_handle_t inproc_device;
     data_size_t  info_size;
     /* VARARG(machines,ushorts); */
-    char __pad_36[4];
 };
 
 
@@ -1255,11 +1202,10 @@ struct get_process_info_reply
     timeout_t    end_time;
     unsigned int session_id;
     int          exit_code;
-    unsigned short priority;
-    unsigned short base_priority;
-    unsigned short disable_boost;
+    int          priority;
     unsigned short machine;
     /* VARARG(image,pe_image_info); */
+    char __pad_62[2];
 };
 
 
@@ -1318,23 +1264,19 @@ struct set_process_info_request
 {
     struct request_header __header;
     obj_handle_t handle;
-    affinity_t   affinity;
+    int          mask;
     int          priority;
-    int          base_priority;
-    int          disable_boost;
+    affinity_t   affinity;
     obj_handle_t token;
-    int          mask;
-    char __pad_44[4];
+    char __pad_36[4];
 };
 struct set_process_info_reply
 {
     struct reply_header __header;
 };
-#define SET_PROCESS_INFO_PRIORITY      0x01
-#define SET_PROCESS_INFO_BASE_PRIORITY 0x02
-#define SET_PROCESS_INFO_DISABLE_BOOST 0x04
-#define SET_PROCESS_INFO_AFFINITY      0x08
-#define SET_PROCESS_INFO_TOKEN         0x10
+#define SET_PROCESS_INFO_PRIORITY 0x01
+#define SET_PROCESS_INFO_AFFINITY 0x02
+#define SET_PROCESS_INFO_TOKEN    0x04
 
 
 
@@ -1355,16 +1297,14 @@ struct get_thread_info_reply
     affinity_t   affinity;
     int          exit_code;
     int          priority;
-    int          base_priority;
+    int          last;
     int          suspend_count;
     unsigned int flags;
     data_size_t  desc_len;
     /* VARARG(desc,unicode_str); */
 };
-#define GET_THREAD_INFO_FLAG_DBG_HIDDEN    0x01
-#define GET_THREAD_INFO_FLAG_TERMINATED    0x02
-#define GET_THREAD_INFO_FLAG_LAST          0x04
-#define GET_THREAD_INFO_FLAG_DISABLE_BOOST 0x08
+#define GET_THREAD_INFO_FLAG_DBG_HIDDEN 0x01
+#define GET_THREAD_INFO_FLAG_TERMINATED 0x02
 
 
 
@@ -1388,28 +1328,24 @@ struct set_thread_info_request
 {
     struct request_header __header;
     obj_handle_t handle;
+    int          mask;
     int          priority;
-    int          base_priority;
     affinity_t   affinity;
     client_ptr_t entry_point;
     obj_handle_t token;
-    int          disable_boost;
-    unsigned int mask;
     /* VARARG(desc,unicode_str); */
-    char __pad_52[4];
+    char __pad_44[4];
 };
 struct set_thread_info_reply
 {
     struct reply_header __header;
 };
-#define SET_THREAD_INFO_PRIORITY        0x01
-#define SET_THREAD_INFO_BASE_PRIORITY   0x02
-#define SET_THREAD_INFO_AFFINITY        0x04
-#define SET_THREAD_INFO_TOKEN           0x08
-#define SET_THREAD_INFO_ENTRYPOINT      0x10
-#define SET_THREAD_INFO_DESCRIPTION     0x20
-#define SET_THREAD_INFO_DBG_HIDDEN      0x40
-#define SET_THREAD_INFO_DISABLE_BOOST   0x80
+#define SET_THREAD_INFO_PRIORITY    0x01
+#define SET_THREAD_INFO_AFFINITY    0x02
+#define SET_THREAD_INFO_TOKEN       0x04
+#define SET_THREAD_INFO_ENTRYPOINT  0x08
+#define SET_THREAD_INFO_DESCRIPTION 0x10
+#define SET_THREAD_INFO_DBG_HIDDEN  0x20
 
 
 
@@ -1445,9 +1381,7 @@ struct queue_apc_request
 {
     struct request_header __header;
     obj_handle_t handle;
-    obj_handle_t reserve_handle;
     /* VARARG(call,apc_call); */
-    char __pad_20[4];
 };
 struct queue_apc_reply
 {
@@ -2210,11 +2144,10 @@ struct get_mapping_info_reply
     mem_size_t   size;
     unsigned int flags;
     obj_handle_t shared_file;
-    data_size_t  name_len;
     data_size_t  total;
     /* VARARG(image,pe_image_info); */
-    /* VARARG(name,unicode_str,name_len); */
-    /* VARARG(exp_name,string); */
+    /* VARARG(name,unicode_str); */
+    char __pad_28[4];
 };
 
 
@@ -2255,10 +2188,9 @@ struct map_image_view_request
     obj_handle_t mapping;
     client_ptr_t base;
     mem_size_t   size;
-    mem_size_t   offset;
     unsigned int entry;
     unsigned short machine;
-    char __pad_46[2];
+    char __pad_38[2];
 };
 struct map_image_view_reply
 {
@@ -2547,8 +2479,6 @@ struct write_process_memory_request
 struct write_process_memory_reply
 {
     struct reply_header __header;
-    data_size_t written;
-    char __pad_12[4];
 };
 
 
@@ -2888,6 +2818,24 @@ struct set_thread_context_reply
 
 
 
+struct get_selector_entry_request
+{
+    struct request_header __header;
+    obj_handle_t  handle;
+    int           entry;
+    char __pad_20[4];
+};
+struct get_selector_entry_reply
+{
+    struct reply_header __header;
+    unsigned int  base;
+    unsigned int  limit;
+    unsigned char flags;
+    char __pad_17[7];
+};
+
+
+
 struct add_atom_request
 {
     struct request_header __header;
@@ -2947,36 +2895,6 @@ struct get_atom_information_reply
 
 
 
-struct add_user_atom_request
-{
-    struct request_header __header;
-    /* VARARG(name,unicode_str); */
-    char __pad_12[4];
-};
-struct add_user_atom_reply
-{
-    struct reply_header __header;
-    atom_t        atom;
-    char __pad_12[4];
-};
-
-
-
-struct get_user_atom_name_request
-{
-    struct request_header __header;
-    atom_t       atom;
-};
-struct get_user_atom_name_reply
-{
-    struct reply_header __header;
-    data_size_t  total;
-    /* VARARG(name,unicode_str); */
-    char __pad_12[4];
-};
-
-
-
 struct get_msg_queue_handle_request
 {
     struct request_header __header;
@@ -2986,7 +2904,7 @@ struct get_msg_queue_handle_reply
 {
     struct reply_header __header;
     obj_handle_t handle;
-    obj_handle_t idle_event;
+    char __pad_12[4];
 };
 
 
@@ -3021,7 +2939,7 @@ struct set_queue_mask_request
     struct request_header __header;
     unsigned int wake_mask;
     unsigned int changed_mask;
-    int          poll_events;
+    int          skip_wait;
 };
 struct set_queue_mask_reply
 {
@@ -3316,8 +3234,6 @@ struct cancel_async_request
 struct cancel_async_reply
 {
     struct reply_header __header;
-    obj_handle_t cancel_handle;
-    char __pad_12[4];
 };
 
 
@@ -3469,13 +3385,12 @@ struct create_window_request
     user_handle_t  parent;
     user_handle_t  owner;
     atom_t         atom;
-    mod_handle_t   class_instance;
     mod_handle_t   instance;
     unsigned int   dpi_context;
     unsigned int   style;
     unsigned int   ex_style;
     /* VARARG(class,unicode_str); */
-    char __pad_52[4];
+    char __pad_44[4];
 };
 struct create_window_reply
 {
@@ -3485,6 +3400,8 @@ struct create_window_reply
     user_handle_t  owner;
     int            extra;
     client_ptr_t   class_ptr;
+    unsigned int   dpi_context;
+    char __pad_36[4];
 };
 
 
@@ -3535,48 +3452,54 @@ struct get_window_info_request
 {
     struct request_header __header;
     user_handle_t  handle;
-    int            offset;
-    data_size_t    size;
 };
 struct get_window_info_reply
 {
     struct reply_header __header;
+    user_handle_t  full_handle;
     user_handle_t  last_active;
+    process_id_t   pid;
+    thread_id_t    tid;
+    atom_t         atom;
     int            is_unicode;
-    lparam_t       info;
+    unsigned int   dpi_context;
+    char __pad_36[4];
 };
 
 
 
-struct init_window_info_request
+struct set_window_info_request
 {
     struct request_header __header;
+    unsigned short flags;
+    short int      is_unicode;
     user_handle_t  handle;
     unsigned int   style;
     unsigned int   ex_style;
-    short int      is_unicode;
-    char __pad_26[6];
-};
-struct init_window_info_reply
-{
-    struct reply_header __header;
-};
-
-
-
-struct set_window_info_request
-{
-    struct request_header __header;
-    user_handle_t  handle;
-    int            offset;
-    data_size_t    size;
-    lparam_t       new_info;
+    data_size_t    extra_size;
+    mod_handle_t   instance;
+    lparam_t       user_data;
+    lparam_t       extra_value;
+    int            extra_offset;
+    char __pad_60[4];
 };
 struct set_window_info_reply
 {
     struct reply_header __header;
-    lparam_t       old_info;
+    unsigned int   old_style;
+    unsigned int   old_ex_style;
+    mod_handle_t   old_instance;
+    lparam_t       old_user_data;
+    lparam_t       old_extra_value;
+    lparam_t       old_id;
 };
+#define SET_WIN_STYLE     0x01
+#define SET_WIN_EXSTYLE   0x02
+#define SET_WIN_ID        0x04
+#define SET_WIN_INSTANCE  0x08
+#define SET_WIN_USERDATA  0x10
+#define SET_WIN_EXTRA     0x20
+#define SET_WIN_UNICODE   0x40
 
 
 
@@ -3592,6 +3515,8 @@ struct set_parent_reply
     struct reply_header __header;
     user_handle_t  old_parent;
     user_handle_t  full_parent;
+    unsigned int   dpi_context;
+    char __pad_20[4];
 };
 
 
@@ -4288,8 +4213,6 @@ struct set_foreground_window_request
 {
     struct request_header __header;
     user_handle_t  handle;
-    int            internal;
-    char __pad_20[4];
 };
 struct set_foreground_window_reply
 {
@@ -4501,18 +4424,18 @@ struct create_class_request
     atom_t         atom;
     unsigned int   style;
     mod_handle_t   instance;
+    int            extra;
+    int            win_extra;
     client_ptr_t   client_ptr;
-    short int      cls_extra;
-    short int      win_extra;
     data_size_t    name_offset;
     /* VARARG(name,unicode_str); */
+    char __pad_52[4];
 };
 struct create_class_reply
 {
     struct reply_header __header;
-    struct obj_locator locator;
-    atom_t             atom;
-    char __pad_28[4];
+    atom_t         atom;
+    char __pad_12[4];
 };
 
 
@@ -4536,29 +4459,32 @@ struct set_class_info_request
 {
     struct request_header __header;
     user_handle_t  window;
-    int            offset;
-    data_size_t    size;
-    lparam_t       new_info;
+    unsigned int   flags;
+    atom_t         atom;
+    unsigned int   style;
+    int            win_extra;
+    mod_handle_t   instance;
+    int            extra_offset;
+    data_size_t    extra_size;
+    lparam_t       extra_value;
 };
 struct set_class_info_reply
 {
     struct reply_header __header;
-    lparam_t       old_info;
-};
-
-
-struct get_class_info_request
-{
-    struct request_header __header;
-    user_handle_t  window;
-    int            offset;
-    data_size_t    size;
-};
-struct get_class_info_reply
-{
-    struct reply_header __header;
-    lparam_t       info;
+    atom_t         old_atom;
+    atom_t         base_atom;
+    mod_handle_t   old_instance;
+    lparam_t       old_extra_value;
+    unsigned int   old_style;
+    int            old_extra;
+    int            old_win_extra;
+    char __pad_44[4];
 };
+#define SET_CLASS_ATOM      0x0001
+#define SET_CLASS_STYLE     0x0002
+#define SET_CLASS_WINEXTRA  0x0004
+#define SET_CLASS_INSTANCE  0x0008
+#define SET_CLASS_EXTRA     0x0010
 
 
 
@@ -5438,11 +5364,9 @@ struct get_token_info_reply
     unsigned int   session_id;
     int            primary;
     int            impersonation_level;
-    int            elevation_type;
-    int            is_elevated;
+    int            elevation;
     int            group_count;
     int            privilege_count;
-    char __pad_52[4];
 };
 
 
@@ -5687,8 +5611,7 @@ struct set_window_layered_info_reply
 struct alloc_user_handle_request
 {
     struct request_header __header;
-    unsigned short type;
-    char __pad_14[2];
+    char __pad_12[4];
 };
 struct alloc_user_handle_reply
 {
@@ -5702,10 +5625,7 @@ struct alloc_user_handle_reply
 struct free_user_handle_request
 {
     struct request_header __header;
-    unsigned short type;
-    char __pad_14[2];
     user_handle_t  handle;
-    char __pad_20[4];
 };
 struct free_user_handle_reply
 {
@@ -5929,22 +5849,6 @@ struct resume_process_reply
 };
 
 
-struct get_next_process_request
-{
-    struct request_header __header;
-    obj_handle_t last;
-    unsigned int access;
-    unsigned int attributes;
-    unsigned int flags;
-    char __pad_28[4];
-};
-struct get_next_process_reply
-{
-    struct reply_header __header;
-    obj_handle_t handle;
-    char __pad_12[4];
-};
-
 
 struct get_next_thread_request
 {
@@ -5979,186 +5883,6 @@ struct set_keyboard_repeat_reply
 };
 
 
-enum inproc_sync_type
-{
-    INPROC_SYNC_UNKNOWN   = 0,
-    INPROC_SYNC_INTERNAL  = 1,
-    INPROC_SYNC_EVENT     = 2,
-    INPROC_SYNC_MUTEX     = 3,
-    INPROC_SYNC_SEMAPHORE = 4,
-};
-
-
-struct get_inproc_sync_fd_request
-{
-    struct request_header __header;
-    obj_handle_t handle;
-};
-struct get_inproc_sync_fd_reply
-{
-    struct reply_header __header;
-    int           type;
-    unsigned int access;
-};
-
-
-
-struct get_inproc_alert_fd_request
-{
-    struct request_header __header;
-    char __pad_12[4];
-};
-struct get_inproc_alert_fd_reply
-{
-    struct reply_header __header;
-    obj_handle_t handle;
-    char __pad_12[4];
-};
-
-
-
-struct d3dkmt_object_create_request
-{
-    struct request_header __header;
-    unsigned int        type;
-    int                 fd;
-    unsigned int        value;
-    /* VARARG(runtime,bytes); */
-};
-struct d3dkmt_object_create_reply
-{
-    struct reply_header __header;
-    d3dkmt_handle_t     global;
-    obj_handle_t        handle;
-};
-
-
-
-struct d3dkmt_object_update_request
-{
-    struct request_header __header;
-    unsigned int        type;
-    d3dkmt_handle_t     global;
-    /* VARARG(runtime,bytes); */
-    char __pad_20[4];
-};
-struct d3dkmt_object_update_reply
-{
-    struct reply_header __header;
-};
-
-
-
-struct d3dkmt_object_query_request
-{
-    struct request_header __header;
-    unsigned int        type;
-    d3dkmt_handle_t     global;
-    obj_handle_t        handle;
-};
-struct d3dkmt_object_query_reply
-{
-    struct reply_header __header;
-    data_size_t         runtime_size;
-    char __pad_12[4];
-};
-
-
-
-struct d3dkmt_object_open_request
-{
-    struct request_header __header;
-    unsigned int        type;
-    d3dkmt_handle_t     global;
-    obj_handle_t        handle;
-};
-struct d3dkmt_object_open_reply
-{
-    struct reply_header __header;
-    d3dkmt_handle_t     global;
-    obj_handle_t        handle;
-    data_size_t         runtime_size;
-    /* VARARG(runtime,bytes); */
-    char __pad_20[4];
-};
-
-
-
-struct d3dkmt_share_objects_request
-{
-    struct request_header __header;
-    d3dkmt_handle_t     resource;
-    d3dkmt_handle_t     mutex;
-    d3dkmt_handle_t     sync;
-    unsigned int        access;
-    /* VARARG(objattr,object_attributes); */
-    char __pad_28[4];
-};
-struct d3dkmt_share_objects_reply
-{
-    struct reply_header __header;
-    obj_handle_t        handle;
-    char __pad_12[4];
-};
-
-
-
-struct d3dkmt_object_open_name_request
-{
-    struct request_header __header;
-    unsigned int        type;
-    unsigned int        access;
-    unsigned int        attributes;
-    obj_handle_t        rootdir;
-    /* VARARG(name,unicode_str); */
-    char __pad_28[4];
-};
-struct d3dkmt_object_open_name_reply
-{
-    struct reply_header __header;
-    obj_handle_t        handle;
-    char __pad_12[4];
-};
-
-
-
-struct d3dkmt_mutex_acquire_request
-{
-    struct request_header __header;
-    d3dkmt_handle_t     mutex;
-    unsigned int        key_value;
-    obj_handle_t        wait_handle;
-    unsigned int        wait_status;
-    char __pad_28[4];
-};
-struct d3dkmt_mutex_acquire_reply
-{
-    struct reply_header __header;
-    unsigned __int64    fence_value;
-    data_size_t         runtime_size;
-    obj_handle_t        wait_handle;
-    /* VARARG(runtime,bytes); */
-};
-
-
-
-struct d3dkmt_mutex_release_request
-{
-    struct request_header __header;
-    d3dkmt_handle_t     mutex;
-    int                 abandon;
-    unsigned int        key_value;
-    unsigned __int64    fence_value;
-    data_size_t         runtime_size;
-    /* VARARG(runtime,bytes); */
-    char __pad_36[4];
-};
-struct d3dkmt_mutex_release_reply
-{
-    struct reply_header __header;
-};
-
-
 enum request
 {
     REQ_new_process,
@@ -6268,12 +5992,11 @@ enum request
     REQ_get_timer_info,
     REQ_get_thread_context,
     REQ_set_thread_context,
+    REQ_get_selector_entry,
     REQ_add_atom,
     REQ_delete_atom,
     REQ_find_atom,
     REQ_get_atom_information,
-    REQ_add_user_atom,
-    REQ_get_user_atom_name,
     REQ_get_msg_queue_handle,
     REQ_get_msg_queue,
     REQ_set_queue_fd,
@@ -6308,7 +6031,6 @@ enum request
     REQ_get_desktop_window,
     REQ_set_window_owner,
     REQ_get_window_info,
-    REQ_init_window_info,
     REQ_set_window_info,
     REQ_set_parent,
     REQ_get_window_parents,
@@ -6367,7 +6089,6 @@ enum request
     REQ_create_class,
     REQ_destroy_class,
     REQ_set_class_info,
-    REQ_get_class_info,
     REQ_open_clipboard,
     REQ_close_clipboard,
     REQ_empty_clipboard,
@@ -6454,19 +6175,8 @@ enum request
     REQ_terminate_job,
     REQ_suspend_process,
     REQ_resume_process,
-    REQ_get_next_process,
     REQ_get_next_thread,
     REQ_set_keyboard_repeat,
-    REQ_get_inproc_sync_fd,
-    REQ_get_inproc_alert_fd,
-    REQ_d3dkmt_object_create,
-    REQ_d3dkmt_object_update,
-    REQ_d3dkmt_object_query,
-    REQ_d3dkmt_object_open,
-    REQ_d3dkmt_share_objects,
-    REQ_d3dkmt_object_open_name,
-    REQ_d3dkmt_mutex_acquire,
-    REQ_d3dkmt_mutex_release,
     REQ_NB_REQUESTS
 };
 
@@ -6581,12 +6291,11 @@ union generic_request
     struct get_timer_info_request get_timer_info_request;
     struct get_thread_context_request get_thread_context_request;
     struct set_thread_context_request set_thread_context_request;
+    struct get_selector_entry_request get_selector_entry_request;
     struct add_atom_request add_atom_request;
     struct delete_atom_request delete_atom_request;
     struct find_atom_request find_atom_request;
     struct get_atom_information_request get_atom_information_request;
-    struct add_user_atom_request add_user_atom_request;
-    struct get_user_atom_name_request get_user_atom_name_request;
     struct get_msg_queue_handle_request get_msg_queue_handle_request;
     struct get_msg_queue_request get_msg_queue_request;
     struct set_queue_fd_request set_queue_fd_request;
@@ -6621,7 +6330,6 @@ union generic_request
     struct get_desktop_window_request get_desktop_window_request;
     struct set_window_owner_request set_window_owner_request;
     struct get_window_info_request get_window_info_request;
-    struct init_window_info_request init_window_info_request;
     struct set_window_info_request set_window_info_request;
     struct set_parent_request set_parent_request;
     struct get_window_parents_request get_window_parents_request;
@@ -6680,7 +6388,6 @@ union generic_request
     struct create_class_request create_class_request;
     struct destroy_class_request destroy_class_request;
     struct set_class_info_request set_class_info_request;
-    struct get_class_info_request get_class_info_request;
     struct open_clipboard_request open_clipboard_request;
     struct close_clipboard_request close_clipboard_request;
     struct empty_clipboard_request empty_clipboard_request;
@@ -6767,19 +6474,8 @@ union generic_request
     struct terminate_job_request terminate_job_request;
     struct suspend_process_request suspend_process_request;
     struct resume_process_request resume_process_request;
-    struct get_next_process_request get_next_process_request;
     struct get_next_thread_request get_next_thread_request;
     struct set_keyboard_repeat_request set_keyboard_repeat_request;
-    struct get_inproc_sync_fd_request get_inproc_sync_fd_request;
-    struct get_inproc_alert_fd_request get_inproc_alert_fd_request;
-    struct d3dkmt_object_create_request d3dkmt_object_create_request;
-    struct d3dkmt_object_update_request d3dkmt_object_update_request;
-    struct d3dkmt_object_query_request d3dkmt_object_query_request;
-    struct d3dkmt_object_open_request d3dkmt_object_open_request;
-    struct d3dkmt_share_objects_request d3dkmt_share_objects_request;
-    struct d3dkmt_object_open_name_request d3dkmt_object_open_name_request;
-    struct d3dkmt_mutex_acquire_request d3dkmt_mutex_acquire_request;
-    struct d3dkmt_mutex_release_request d3dkmt_mutex_release_request;
 };
 union generic_reply
 {
@@ -6892,12 +6588,11 @@ union generic_reply
     struct get_timer_info_reply get_timer_info_reply;
     struct get_thread_context_reply get_thread_context_reply;
     struct set_thread_context_reply set_thread_context_reply;
+    struct get_selector_entry_reply get_selector_entry_reply;
     struct add_atom_reply add_atom_reply;
     struct delete_atom_reply delete_atom_reply;
     struct find_atom_reply find_atom_reply;
     struct get_atom_information_reply get_atom_information_reply;
-    struct add_user_atom_reply add_user_atom_reply;
-    struct get_user_atom_name_reply get_user_atom_name_reply;
     struct get_msg_queue_handle_reply get_msg_queue_handle_reply;
     struct get_msg_queue_reply get_msg_queue_reply;
     struct set_queue_fd_reply set_queue_fd_reply;
@@ -6932,7 +6627,6 @@ union generic_reply
     struct get_desktop_window_reply get_desktop_window_reply;
     struct set_window_owner_reply set_window_owner_reply;
     struct get_window_info_reply get_window_info_reply;
-    struct init_window_info_reply init_window_info_reply;
     struct set_window_info_reply set_window_info_reply;
     struct set_parent_reply set_parent_reply;
     struct get_window_parents_reply get_window_parents_reply;
@@ -6991,7 +6685,6 @@ union generic_reply
     struct create_class_reply create_class_reply;
     struct destroy_class_reply destroy_class_reply;
     struct set_class_info_reply set_class_info_reply;
-    struct get_class_info_reply get_class_info_reply;
     struct open_clipboard_reply open_clipboard_reply;
     struct close_clipboard_reply close_clipboard_reply;
     struct empty_clipboard_reply empty_clipboard_reply;
@@ -7078,21 +6771,10 @@ union generic_reply
     struct terminate_job_reply terminate_job_reply;
     struct suspend_process_reply suspend_process_reply;
     struct resume_process_reply resume_process_reply;
-    struct get_next_process_reply get_next_process_reply;
     struct get_next_thread_reply get_next_thread_reply;
     struct set_keyboard_repeat_reply set_keyboard_repeat_reply;
-    struct get_inproc_sync_fd_reply get_inproc_sync_fd_reply;
-    struct get_inproc_alert_fd_reply get_inproc_alert_fd_reply;
-    struct d3dkmt_object_create_reply d3dkmt_object_create_reply;
-    struct d3dkmt_object_update_reply d3dkmt_object_update_reply;
-    struct d3dkmt_object_query_reply d3dkmt_object_query_reply;
-    struct d3dkmt_object_open_reply d3dkmt_object_open_reply;
-    struct d3dkmt_share_objects_reply d3dkmt_share_objects_reply;
-    struct d3dkmt_object_open_name_reply d3dkmt_object_open_name_reply;
-    struct d3dkmt_mutex_acquire_reply d3dkmt_mutex_acquire_reply;
-    struct d3dkmt_mutex_release_reply d3dkmt_mutex_release_reply;
-};
-
-#define SERVER_PROTOCOL_VERSION 930
+};
+
+#define SERVER_PROTOCOL_VERSION 855
 
 #endif /* __WINE_WINE_SERVER_PROTOCOL_H */
