diff --git a/include/wine/winuser16.h b/include/wine/winuser16.h
index 95835defc8fc337c40877d996103e4d4b596f3ca..90e80c7fdca152b62abcce78dd5ac6d11420e945 100644
--- a/include/wine/winuser16.h
+++ b/include/wine/winuser16.h
@@ -22,7 +22,7 @@
 #include <wine/wingdi16.h> /* wingdi.h needed for COLORREF */
 #include <winuser.h> /* winuser.h needed for MSGBOXCALLBACK */
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 typedef struct tagCOMSTAT16
 {
@@ -556,7 +556,7 @@ typedef struct tagCOPYDATASTRUCT16 {
 #define DRAG_PRINT 0x544E5250
 #define DRAG_FILE  0x454C4946
 
-#pragma pack(pop)
+#include <poppack.h>
 
 /* WM_COMMNOTIFY flags */
 #define CN_RECEIVE	0x0001
