diff --git a/include/wininet.h b/include/wininet.h
index 0537585adf8fe98c64a8850731a6eca444ea7dbf..0de2b8308120fb3a152f2ff825cfd997280ee523 100644
--- a/include/wininet.h
+++ b/include/wininet.h
@@ -1242,7 +1242,6 @@ BOOLAPI GopherGetAttributeW(HINTERNET ,LPCWSTR ,LPCWSTR ,LPBYTE ,
 #define HTTP_STATUS_NOT_MODIFIED        304
 #define HTTP_STATUS_USE_PROXY           305
 #define HTTP_STATUS_REDIRECT_KEEP_VERB  307
-#define HTTP_STATUS_PERMANENT_REDIRECT  308
 #define HTTP_STATUS_BAD_REQUEST         400
 #define HTTP_STATUS_DENIED              401
 #define HTTP_STATUS_PAYMENT_REQ         402
