diff --git a/include/winnt.h b/include/winnt.h
index bd78fc9ed4d5882eec2e7530346247c41b96a561..0c6fbbd803457a98b7d32b501739900bd1c13a34 100644
--- a/include/winnt.h
+++ b/include/winnt.h
@@ -83,6 +83,16 @@ extern "C" {
 # endif
 #endif
 
+#ifndef DECLSPEC_ALLOCATE
+# if __has_declspec_attribute(allocate) && !defined(MIDL_PASS)
+#  define DECLSPEC_ALLOCATE(x) __declspec(allocate(x))
+# elif defined(__GNUC__)
+#  define DECLSPEC_ALLOCATE(x) __attribute__((section(x)))
+# else
+#  define DECLSPEC_ALLOCATE(x)
+# endif
+#endif
+
 #ifndef DECLSPEC_NOTHROW
 # ifdef __GNUC__
 #  define DECLSPEC_NOTHROW __attribute__((nothrow))
@@ -401,9 +411,7 @@ extern "C" {
 
 /* Compile time assertion */
 
-#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 202311L)
-#define C_ASSERT(e) static_assert(e, #e)
-#elif defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)
+#if defined(__STDC_VERSION__) && (__STDC_VERSION__ >= 201112L)
 #define C_ASSERT(e) _Static_assert(e, #e)
 #else
 #define C_ASSERT(e) extern void __C_ASSERT__(int [(e)?1:-1])
@@ -471,14 +479,14 @@ typedef VOID           *PVOID64;
 typedef BYTE            BOOLEAN,    *PBOOLEAN;
 typedef char            CHAR,       *PCHAR;
 typedef short           SHORT,      *PSHORT;
-#if !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES) && !defined(WINE_UNIX_LIB)
+#if !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES)
 typedef long            LONG,       *PLONG;
 #else
 typedef int             LONG,       *PLONG;
 #endif
 
 /* Some systems might have wchar_t, but we really need 16 bit characters */
-#if defined(WINE_UNICODE_NATIVE) || defined(__MINGW32__) || defined(_MSC_VER)
+#if defined(WINE_UNICODE_NATIVE)
 typedef wchar_t         WCHAR;
 #elif __cpp_unicode_literals >= 200710
 typedef char16_t        WCHAR;
@@ -496,19 +504,21 @@ typedef ULONG           UCSCHAR;
 #ifndef _ULONGLONG_
 # define _ULONGLONG_
 # ifdef _MSC_VER
-typedef signed __int64   LONGLONG;
-typedef unsigned __int64 ULONGLONG;
+typedef signed __int64   LONGLONG,  *PLONGLONG;
+typedef unsigned __int64 ULONGLONG, *PULONGLONG;
 # else
-typedef signed __int64   DECLSPEC_ALIGN(8) LONGLONG;
-typedef unsigned __int64 DECLSPEC_ALIGN(8) ULONGLONG;
+typedef signed __int64   DECLSPEC_ALIGN(8) LONGLONG,   *PLONGLONG;
+typedef unsigned __int64 DECLSPEC_ALIGN(8) ULONGLONG,  *PULONGLONG;
 # endif
-typedef LONGLONG *PLONGLONG;
-typedef ULONGLONG *PULONGLONG;
 #endif
 
 #ifndef _DWORDLONG_
 # define _DWORDLONG_
+# ifdef _MSC_VER
 typedef ULONGLONG DWORDLONG, *PDWORDLONG;
+# else
+typedef ULONGLONG   DECLSPEC_ALIGN(8) DWORDLONG,   *PDWORDLONG;
+# endif
 #endif
 
 /* ANSI string types */
@@ -607,7 +617,7 @@ typedef DWORD FLONG;
 
 /* Macro to deal with LP64 <=> LLP64 differences in numeric constants with 'l' modifier */
 #ifndef __MSABI_LONG
-#if !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES) && !defined(WINE_UNIX_LIB)
+#if !defined(__LP64__) && !defined(WINE_NO_LONG_TYPES)
 #  define __MSABI_LONG(x)         x ## l
 # else
 #  define __MSABI_LONG(x)         x
@@ -734,11 +744,10 @@ typedef DWORD FLONG;
 #define PROCESSOR_OPTIL          18767
 
 #ifdef _WIN64
-#define MAXIMUM_PROC_PER_GROUP   64
+#define MAXIMUM_PROCESSORS       64
 #else
-#define MAXIMUM_PROC_PER_GROUP   32
+#define MAXIMUM_PROCESSORS       32
 #endif
-#define MAXIMUM_PROCESSORS       MAXIMUM_PROC_PER_GROUP
 
 typedef struct _MEMORY_BASIC_INFORMATION
 {
@@ -1053,95 +1062,67 @@ typedef enum _HEAP_INFORMATION_CLASS {
 } HEAP_INFORMATION_CLASS;
 
 /* Processor feature flags.  */
-#define PF_FLOATING_POINT_PRECISION_ERRATA          0
-#define PF_FLOATING_POINT_EMULATED                  1
-#define PF_COMPARE_EXCHANGE_DOUBLE                  2
-#define PF_MMX_INSTRUCTIONS_AVAILABLE               3
-#define PF_PPC_MOVEMEM_64BIT_OK                     4
-#define PF_ALPHA_BYTE_INSTRUCTIONS                  5
-#define PF_XMMI_INSTRUCTIONS_AVAILABLE              6
-#define PF_3DNOW_INSTRUCTIONS_AVAILABLE             7
-#define PF_RDTSC_INSTRUCTION_AVAILABLE              8
-#define PF_PAE_ENABLED                              9
-#define PF_XMMI64_INSTRUCTIONS_AVAILABLE            10
-#define PF_SSE_DAZ_MODE_AVAILABLE                   11
-#define PF_NX_ENABLED                               12
-#define PF_SSE3_INSTRUCTIONS_AVAILABLE              13
-#define PF_COMPARE_EXCHANGE128                      14
-#define PF_COMPARE64_EXCHANGE128                    15
-#define PF_CHANNELS_ENABLED                         16
-#define PF_XSAVE_ENABLED                            17
-#define PF_ARM_VFP_32_REGISTERS_AVAILABLE           18
-#define PF_ARM_NEON_INSTRUCTIONS_AVAILABLE          19
-#define PF_SECOND_LEVEL_ADDRESS_TRANSLATION         20
-#define PF_VIRT_FIRMWARE_ENABLED                    21
-#define PF_RDWRFSGSBASE_AVAILABLE                   22
-#define PF_FASTFAIL_AVAILABLE                       23
-#define PF_ARM_DIVIDE_INSTRUCTION_AVAILABLE         24
-#define PF_ARM_64BIT_LOADSTORE_ATOMIC               25
-#define PF_ARM_EXTERNAL_CACHE_AVAILABLE             26
-#define PF_ARM_FMAC_INSTRUCTIONS_AVAILABLE          27
-#define PF_RDRAND_INSTRUCTION_AVAILABLE             28
-#define PF_ARM_V8_INSTRUCTIONS_AVAILABLE            29
-#define PF_ARM_V8_CRYPTO_INSTRUCTIONS_AVAILABLE     30
-#define PF_ARM_V8_CRC32_INSTRUCTIONS_AVAILABLE      31
-#define PF_RDTSCP_INSTRUCTION_AVAILABLE             32
-#define PF_RDPID_INSTRUCTION_AVAILABLE              33
-#define PF_ARM_V81_ATOMIC_INSTRUCTIONS_AVAILABLE    34
-#define PF_MONITORX_INSTRUCTION_AVAILABLE           35
-#define PF_SSSE3_INSTRUCTIONS_AVAILABLE             36
-#define PF_SSE4_1_INSTRUCTIONS_AVAILABLE            37
-#define PF_SSE4_2_INSTRUCTIONS_AVAILABLE            38
-#define PF_AVX_INSTRUCTIONS_AVAILABLE               39
-#define PF_AVX2_INSTRUCTIONS_AVAILABLE              40
-#define PF_AVX512F_INSTRUCTIONS_AVAILABLE           41
-#define PF_ERMS_AVAILABLE                           42
-#define PF_ARM_V82_DP_INSTRUCTIONS_AVAILABLE        43
-#define PF_ARM_V83_JSCVT_INSTRUCTIONS_AVAILABLE     44
-#define PF_ARM_V83_LRCPC_INSTRUCTIONS_AVAILABLE     45
-#define PF_ARM_SVE_INSTRUCTIONS_AVAILABLE           46
-#define PF_ARM_SVE2_INSTRUCTIONS_AVAILABLE          47
-#define PF_ARM_SVE2_1_INSTRUCTIONS_AVAILABLE        48
-#define PF_ARM_SVE_AES_INSTRUCTIONS_AVAILABLE       49
-#define PF_ARM_SVE_PMULL128_INSTRUCTIONS_AVAILABLE  50
-#define PF_ARM_SVE_BITPERM_INSTRUCTIONS_AVAILABLE   51
-#define PF_ARM_SVE_BF16_INSTRUCTIONS_AVAILABLE      52
-#define PF_ARM_SVE_EBF16_INSTRUCTIONS_AVAILABLE     53
-#define PF_ARM_SVE_B16B16_INSTRUCTIONS_AVAILABLE    54
-#define PF_ARM_SVE_SHA3_INSTRUCTIONS_AVAILABLE      55
-#define PF_ARM_SVE_SM4_INSTRUCTIONS_AVAILABLE       56
-#define PF_ARM_SVE_I8MM_INSTRUCTIONS_AVAILABLE      57
-#define PF_ARM_SVE_F32MM_INSTRUCTIONS_AVAILABLE     58
-#define PF_ARM_SVE_F64MM_INSTRUCTIONS_AVAILABLE     59
-#define PF_BMI2_INSTRUCTIONS_AVAILABLE              60
-#define PF_MOVDIR64B_INSTRUCTION_AVAILABLE          61
-#define PF_ARM_LSE2_AVAILABLE                       62
-#define PF_RESERVED_FEATURE                         63
-#define PF_ARM_SHA3_INSTRUCTIONS_AVAILABLE          64
-#define PF_ARM_SHA512_INSTRUCTIONS_AVAILABLE        65
-#define PF_ARM_V82_I8MM_INSTRUCTIONS_AVAILABLE      66
-#define PF_ARM_V82_FP16_INSTRUCTIONS_AVAILABLE      67
-#define PF_ARM_V86_BF16_INSTRUCTIONS_AVAILABLE      68
-#define PF_ARM_V86_EBF16_INSTRUCTIONS_AVAILABLE     69
-#define PF_ARM_SME_INSTRUCTIONS_AVAILABLE           70
-#define PF_ARM_SME2_INSTRUCTIONS_AVAILABLE          71
-#define PF_ARM_SME2_1_INSTRUCTIONS_AVAILABLE        72
-#define PF_ARM_SME2_2_INSTRUCTIONS_AVAILABLE        73
-#define PF_ARM_SME_AES_INSTRUCTIONS_AVAILABLE       74
-#define PF_ARM_SME_SBITPERM_INSTRUCTIONS_AVAILABLE  75
-#define PF_ARM_SME_SF8MM4_INSTRUCTIONS_AVAILABLE    76
-#define PF_ARM_SME_SF8MM8_INSTRUCTIONS_AVAILABLE    77
-#define PF_ARM_SME_SF8DP2_INSTRUCTIONS_AVAILABLE    78
-#define PF_ARM_SME_SF8DP4_INSTRUCTIONS_AVAILABLE    79
-#define PF_ARM_SME_SF8FMA_INSTRUCTIONS_AVAILABLE    80
-#define PF_ARM_SME_F8F32_INSTRUCTIONS_AVAILABLE     81
-#define PF_ARM_SME_F8F16_INSTRUCTIONS_AVAILABLE     82
-#define PF_ARM_SME_F16F16_INSTRUCTIONS_AVAILABLE    83
-#define PF_ARM_SME_B16B16_INSTRUCTIONS_AVAILABLE    84
-#define PF_ARM_SME_F64F64_INSTRUCTIONS_AVAILABLE    85
-#define PF_ARM_SME_I16I64_INSTRUCTIONS_AVAILABLE    86
-#define PF_ARM_SME_LUTv2_INSTRUCTIONS_AVAILABLE     87
-#define PF_ARM_SME_FA64_INSTRUCTIONS_AVAILABLE      88
+#define PF_FLOATING_POINT_PRECISION_ERRATA	0
+#define PF_FLOATING_POINT_EMULATED		1
+#define PF_COMPARE_EXCHANGE_DOUBLE		2
+#define PF_MMX_INSTRUCTIONS_AVAILABLE		3
+#define PF_PPC_MOVEMEM_64BIT_OK			4
+#define PF_ALPHA_BYTE_INSTRUCTIONS		5
+#define PF_XMMI_INSTRUCTIONS_AVAILABLE		6
+#define PF_3DNOW_INSTRUCTIONS_AVAILABLE		7
+#define PF_RDTSC_INSTRUCTION_AVAILABLE		8
+#define PF_PAE_ENABLED				9
+#define PF_XMMI64_INSTRUCTIONS_AVAILABLE	10
+#define PF_SSE_DAZ_MODE_AVAILABLE		11
+#define PF_NX_ENABLED				12
+#define PF_SSE3_INSTRUCTIONS_AVAILABLE		13
+#define PF_COMPARE_EXCHANGE128			14
+#define PF_COMPARE64_EXCHANGE128		15
+#define PF_CHANNELS_ENABLED			16
+#define PF_XSAVE_ENABLED			17
+#define PF_ARM_VFP_32_REGISTERS_AVAILABLE       18
+#define PF_ARM_NEON_INSTRUCTIONS_AVAILABLE      19
+#define PF_SECOND_LEVEL_ADDRESS_TRANSLATION     20
+#define PF_VIRT_FIRMWARE_ENABLED                21
+#define PF_RDWRFSGSBASE_AVAILABLE               22
+#define PF_FASTFAIL_AVAILABLE                   23
+#define PF_ARM_DIVIDE_INSTRUCTION_AVAILABLE     24
+#define PF_ARM_64BIT_LOADSTORE_ATOMIC           25
+#define PF_ARM_EXTERNAL_CACHE_AVAILABLE         26
+#define PF_ARM_FMAC_INSTRUCTIONS_AVAILABLE      27
+#define PF_RDRAND_INSTRUCTION_AVAILABLE         28
+#define PF_ARM_V8_INSTRUCTIONS_AVAILABLE        29
+#define PF_ARM_V8_CRYPTO_INSTRUCTIONS_AVAILABLE 30
+#define PF_ARM_V8_CRC32_INSTRUCTIONS_AVAILABLE  31
+#define PF_RDTSCP_INSTRUCTION_AVAILABLE         32
+#define PF_RDPID_INSTRUCTION_AVAILABLE          33
+#define PF_ARM_V81_ATOMIC_INSTRUCTIONS_AVAILABLE 34
+#define PF_MONITORX_INSTRUCTION_AVAILABLE       35
+#define PF_SSSE3_INSTRUCTIONS_AVAILABLE         36
+#define PF_SSE4_1_INSTRUCTIONS_AVAILABLE        37
+#define PF_SSE4_2_INSTRUCTIONS_AVAILABLE        38
+#define PF_AVX_INSTRUCTIONS_AVAILABLE           39
+#define PF_AVX2_INSTRUCTIONS_AVAILABLE          40
+#define PF_AVX512F_INSTRUCTIONS_AVAILABLE       41
+#define PF_ERMS_AVAILABLE                       42
+#define PF_ARM_V82_DP_INSTRUCTIONS_AVAILABLE    43
+#define PF_ARM_V83_JSCVT_INSTRUCTIONS_AVAILABLE 44
+#define PF_ARM_V83_LRCPC_INSTRUCTIONS_AVAILABLE 45
+#define PF_ARM_SVE_INSTRUCTIONS_AVAILABLE       46
+#define PF_ARM_SVE2_INSTRUCTIONS_AVAILABLE      47
+#define PF_ARM_SVE2_1_INSTRUCTIONS_AVAILABLE    48
+#define PF_ARM_SVE_AES_INSTRUCTIONS_AVAILABLE   49
+#define PF_ARM_SVE_PMULL128_INSTRUCTIONS_AVAILABLE 50
+#define PF_ARM_SVE_BITPERM_INSTRUCTIONS_AVAILABLE 51
+#define PF_ARM_SVE_BF16_INSTRUCTIONS_AVAILABLE  52
+#define PF_ARM_SVE_EBF16_INSTRUCTIONS_AVAILABLE 53
+#define PF_ARM_SVE_B16B16_INSTRUCTIONS_AVAILABLE 54
+#define PF_ARM_SVE_SHA3_INSTRUCTIONS_AVAILABLE  55
+#define PF_ARM_SVE_SM4_INSTRUCTIONS_AVAILABLE   56
+#define PF_ARM_SVE_I8MM_INSTRUCTIONS_AVAILABLE  57
+#define PF_ARM_SVE_F32MM_INSTRUCTIONS_AVAILABLE 58
+#define PF_ARM_SVE_F64MM_INSTRUCTIONS_AVAILABLE 59
+#define PF_BMI2_INSTRUCTIONS_AVAILABLE          60
 
 
 /* Execution state flags */
@@ -1173,7 +1154,7 @@ typedef struct _I386_FLOATING_SAVE_AREA
 
 #define I386_MAXIMUM_SUPPORTED_EXTENSION     512
 
-#pragma pack(push,4)
+#include "pshpack4.h"
 typedef struct _I386_CONTEXT
 {
     DWORD   ContextFlags;  /* 000 */
@@ -1213,7 +1194,7 @@ typedef struct _I386_CONTEXT
 
     BYTE    ExtendedRegisters[I386_MAXIMUM_SUPPORTED_EXTENSION];  /* 0xcc */
 } I386_CONTEXT, WOW64_CONTEXT, *PWOW64_CONTEXT;
-#pragma pack(pop)
+#include "poppack.h"
 
 #define CONTEXT_i386      0x00010000
 #define CONTEXT_i486      0x00010000
@@ -1519,25 +1500,15 @@ typedef struct _XSTATE_CONFIGURATION
     ULONG64 EnabledFeatures;
     ULONG64 EnabledVolatileFeatures;
     ULONG Size;
-    union
-    {
-        ULONG ControlFlags;
-        struct
-        {
-            ULONG OptimizedSave : 1;
-            ULONG CompactionEnabled : 1;
-            ULONG ExtendedFeatureDisable : 1;
-        };
-    };
+    ULONG OptimizedSave:1;
+    ULONG CompactionEnabled:1;
     XSTATE_FEATURE Features[MAXIMUM_XSTATE_FEATURES];
+
     ULONG64 EnabledSupervisorFeatures;
     ULONG64 AlignedFeatures;
     ULONG AllFeatureSize;
     ULONG AllFeatures[MAXIMUM_XSTATE_FEATURES];
     ULONG64 EnabledUserVisibleSupervisorFeatures;
-    ULONG64 ExtendedFeatureDisableFeatures;
-    ULONG AllNonLargeFeatureSize;
-    ULONG Spare;
 } XSTATE_CONFIGURATION, *PXSTATE_CONFIGURATION;
 
 typedef struct _XSAVE_AREA_HEADER
@@ -1734,8 +1705,8 @@ typedef struct _KNONVOLATILE_CONTEXT_POINTERS
 #define CONTEXT_ARM64_INTEGER         (CONTEXT_ARM64 | 0x00000002)
 #define CONTEXT_ARM64_FLOATING_POINT  (CONTEXT_ARM64 | 0x00000004)
 #define CONTEXT_ARM64_DEBUG_REGISTERS (CONTEXT_ARM64 | 0x00000008)
-#define CONTEXT_ARM64_X18       (CONTEXT_ARM64 | 0x00000010)
-#define CONTEXT_ARM64_XSTATE    (CONTEXT_ARM64 | 0x00000020)
+#define CONTEXT_ARM64_X18             (CONTEXT_ARM64 | 0x00000010)
+#define CONTEXT_ARM64_XSTATE          (CONTEXT_ARM64 | 0x00000020)
 #define CONTEXT_ARM64_FEX_YMMSTATE   (CONTEXT_ARM64 | 0x00000040)
 #define CONTEXT_ARM64_FULL (CONTEXT_ARM64_CONTROL | CONTEXT_ARM64_INTEGER | CONTEXT_ARM64_FLOATING_POINT)
 #define CONTEXT_ARM64_ALL  (CONTEXT_ARM64_FULL | CONTEXT_ARM64_DEBUG_REGISTERS | CONTEXT_ARM64_X18)
@@ -2016,6 +1987,7 @@ typedef struct DECLSPEC_ALIGN(16) _ARM64EC_NT_CONTEXT
 #define CONTEXT_INTEGER CONTEXT_ARM64_INTEGER
 #define CONTEXT_FLOATING_POINT CONTEXT_ARM64_FLOATING_POINT
 #define CONTEXT_DEBUG_REGISTERS CONTEXT_ARM64_DEBUG_REGISTERS
+#define CONTEXT_XSTATE CONTEXT_ARM64_XSTATE
 #define CONTEXT_FULL CONTEXT_ARM64_FULL
 #define CONTEXT_ALL CONTEXT_ARM64_ALL
 #define CONTEXT_RET_TO_GUEST CONTEXT_ARM64_RET_TO_GUEST
@@ -2494,7 +2466,7 @@ NTSYSAPI struct _TEB * WINAPI NtCurrentTeb(void) __attribute__((pure));
 # else
 NTSYSAPI struct _TEB * WINAPI NtCurrentTeb(void);
 # endif
-#elif defined(__i386__) && defined(__GNUC__)
+#elif defined(__i386__) && (defined(__GNUC__) || defined(__clang__))
 static FORCEINLINE struct _TEB * WINAPI NtCurrentTeb(void)
 {
     struct _TEB *teb;
@@ -2502,7 +2474,7 @@ static FORCEINLINE struct _TEB * WINAPI NtCurrentTeb(void)
     return teb;
 }
 #elif defined(__i386__) && defined(_MSC_VER)
-DWORD __readfsdword(DWORD);
+unsigned long __readfsdword(unsigned long);
 #pragma intrinsic(__readfsdword)
 static FORCEINLINE struct _TEB * WINAPI NtCurrentTeb(void)
 {
@@ -2624,7 +2596,7 @@ static FORCEINLINE struct _TEB * WINAPI NtCurrentTeb(void)
  * File formats definitions
  */
 
-#pragma pack(push,2)
+#include <pshpack2.h>
 typedef struct _IMAGE_DOS_HEADER {
     WORD  e_magic;      /* 00: MZ Header signature */
     WORD  e_cblp;       /* 02: Bytes on last page of file */
@@ -2646,7 +2618,7 @@ typedef struct _IMAGE_DOS_HEADER {
     WORD  e_res2[10];   /* 28: Reserved words */
     DWORD e_lfanew;     /* 3c: Offset to extended header */
 } IMAGE_DOS_HEADER, *PIMAGE_DOS_HEADER;
-#pragma pack(pop)
+#include <poppack.h>
 
 #define IMAGE_DOS_SIGNATURE    0x5A4D     /* MZ   */
 #define IMAGE_OS2_SIGNATURE    0x454E     /* NE   */
@@ -2659,7 +2631,7 @@ typedef struct _IMAGE_DOS_HEADER {
  * This is the Windows executable (NE) header.
  * the name IMAGE_OS2_HEADER is misleading, but in the SDK this way.
  */
-#pragma pack(push,2)
+#include <pshpack2.h>
 typedef struct
 {
     WORD  ne_magic;             /* 00 NE signature 'NE' */
@@ -2693,9 +2665,9 @@ typedef struct
     WORD  ne_swaparea;          /* 3c Reserved by Microsoft */
     WORD  ne_expver;            /* 3e Expected Windows version number */
 } IMAGE_OS2_HEADER, *PIMAGE_OS2_HEADER;
-#pragma pack(pop)
+#include <poppack.h>
 
-#pragma pack(push,2)
+#include <pshpack2.h>
 typedef struct _IMAGE_VXD_HEADER {
   WORD  e32_magic;
   BYTE  e32_border;
@@ -2749,7 +2721,7 @@ typedef struct _IMAGE_VXD_HEADER {
   WORD  e32_devid;
   WORD  e32_ddkver;
 } IMAGE_VXD_HEADER, *PIMAGE_VXD_HEADER;
-#pragma pack(pop)
+#include <poppack.h>
 
 /* These defines describe the meanings of the bits in the Characteristics
    field */
@@ -3093,7 +3065,7 @@ typedef struct _IMAGE_SECTION_HEADER {
 #define IMAGE_SCN_MEM_READ			0x40000000
 #define IMAGE_SCN_MEM_WRITE			0x80000000
 
-#pragma pack(push,2)
+#include <pshpack2.h>
 
 typedef struct _IMAGE_SYMBOL {
     union {
@@ -3162,7 +3134,7 @@ typedef IMAGE_AUX_SYMBOL *PIMAGE_AUX_SYMBOL;
 
 #define IMAGE_SIZEOF_AUX_SYMBOL 18
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #define IMAGE_SYM_UNDEFINED           (SHORT)0
 #define IMAGE_SYM_ABSOLUTE            (SHORT)-1
@@ -3286,7 +3258,7 @@ typedef struct _IMAGE_IMPORT_BY_NAME {
 	char	Name[1];
 } IMAGE_IMPORT_BY_NAME,*PIMAGE_IMPORT_BY_NAME;
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 /* Import thunk */
 typedef struct _IMAGE_THUNK_DATA64 {
 	union {
@@ -3296,7 +3268,7 @@ typedef struct _IMAGE_THUNK_DATA64 {
 		ULONGLONG AddressOfData;
 	} u1;
 } IMAGE_THUNK_DATA64,*PIMAGE_THUNK_DATA64;
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef struct _IMAGE_THUNK_DATA32 {
 	union {
@@ -3370,7 +3342,7 @@ typedef struct _IMAGE_BASE_RELOCATION
 	/* WORD	TypeOffset[1]; */
 } IMAGE_BASE_RELOCATION,*PIMAGE_BASE_RELOCATION;
 
-#pragma pack(push,2)
+#include <pshpack2.h>
 
 typedef struct _IMAGE_RELOCATION
 {
@@ -3382,7 +3354,7 @@ typedef struct _IMAGE_RELOCATION
     WORD    Type;
 } IMAGE_RELOCATION, *PIMAGE_RELOCATION;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #define IMAGE_SIZEOF_RELOCATION 10
 
@@ -4015,7 +3987,7 @@ typedef struct _IMAGE_DYNAMIC_RELOCATION_TABLE
     DWORD     Size;
 } IMAGE_DYNAMIC_RELOCATION_TABLE, *PIMAGE_DYNAMIC_RELOCATION_TABLE;
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 
 typedef struct _IMAGE_DYNAMIC_RELOCATION32
 {
@@ -4047,7 +4019,7 @@ typedef struct _IMAGE_DYNAMIC_RELOCATION64_V2
     DWORD     Flags;
 } IMAGE_DYNAMIC_RELOCATION64_V2, *PIMAGE_DYNAMIC_RELOCATION64_V2;
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #ifdef _WIN64
 typedef IMAGE_DYNAMIC_RELOCATION64     IMAGE_DYNAMIC_RELOCATION;
@@ -4249,57 +4221,48 @@ typedef enum _TOKEN_ELEVATION_TYPE {
  */
 
 typedef enum _TOKEN_INFORMATION_CLASS {
-    TokenUser = 1,
-    TokenGroups = 2,
-    TokenPrivileges = 3,
-    TokenOwner = 4,
-    TokenPrimaryGroup = 5,
-    TokenDefaultDacl = 6,
-    TokenSource = 7,
-    TokenType = 8,
-    TokenImpersonationLevel = 9,
-    TokenStatistics = 10,
-    TokenRestrictedSids = 11,
-    TokenSessionId = 12,
-    TokenGroupsAndPrivileges = 13,
-    TokenSessionReference = 14,
-    TokenSandBoxInert = 15,
-    TokenAuditPolicy = 16,
-    TokenOrigin = 17,
-    TokenElevationType = 18,
-    TokenLinkedToken = 19,
-    TokenElevation = 20,
-    TokenHasRestrictions = 21,
-    TokenAccessInformation = 22,
-    TokenVirtualizationAllowed = 23,
-    TokenVirtualizationEnabled = 24,
-    TokenIntegrityLevel = 25,
-    TokenUIAccess = 26,
-    TokenMandatoryPolicy = 27,
-    TokenLogonSid = 28,
-    TokenIsAppContainer = 29,
-    TokenCapabilities = 30,
-    TokenAppContainerSid = 31,
-    TokenAppContainerNumber = 32,
-    TokenUserClaimAttributes = 33,
-    TokenDeviceClaimAttributes = 34,
-    TokenRestrictedUserClaimAttributes = 35,
-    TokenRestrictedDeviceClaimAttributes = 36,
-    TokenDeviceGroups = 37,
-    TokenRestrictedDeviceGroups = 38,
-    TokenSecurityAttributes = 39,
-    TokenIsRestricted = 40,
-    TokenProcessTrustLevel = 41,
-    TokenPrivateNameSpace = 42,
-    TokenSingletonAttributes = 43,
-    TokenBnoIsolation = 44,
-    TokenChildProcessFlags = 45,
-    TokenIsLessPrivilegedAppContainer = 46,
-    TokenIsSandboxed = 47,
-    TokenIsAppSilo = 48,
-    TokenLoggingInformation = 49,
-    TokenLearningMode = 50,
-    MaxTokenInfoClass
+  TokenUser = 1,
+  TokenGroups,
+  TokenPrivileges,
+  TokenOwner,
+  TokenPrimaryGroup,
+  TokenDefaultDacl,
+  TokenSource,
+  TokenType,
+  TokenImpersonationLevel,
+  TokenStatistics,
+  TokenRestrictedSids,
+  TokenSessionId,
+  TokenGroupsAndPrivileges,
+  TokenSessionReference,
+  TokenSandBoxInert,
+  TokenAuditPolicy,
+  TokenOrigin,
+  TokenElevationType,
+  TokenLinkedToken,
+  TokenElevation,
+  TokenHasRestrictions,
+  TokenAccessInformation,
+  TokenVirtualizationAllowed,
+  TokenVirtualizationEnabled,
+  TokenIntegrityLevel,
+  TokenUIAccess,
+  TokenMandatoryPolicy,
+  TokenLogonSid,
+  TokenIsAppContainer,
+  TokenCapabilities,
+  TokenAppContainerSid,
+  TokenAppContainerNumber,
+  TokenUserClaimAttributes,
+  TokenDeviceClaimAttributes,
+  TokenRestrictedUserClaimAttributes,
+  TokenRestrictedDeviceClaimAttributes,
+  TokenDeviceGroups,
+  TokenRestrictedDeviceGroups,
+  TokenSecurityAttributes,
+  TokenIsRestricted,
+  TokenProcessTrustLevel,
+  MaxTokenInfoClass
 } TOKEN_INFORMATION_CLASS;
 
 #define DISABLE_MAX_PRIVILEGE        0x1
@@ -4394,7 +4357,7 @@ typedef struct _ACL {
 
 typedef enum _ACL_INFORMATION_CLASS
 {
-  AclRevisionInformation = 1,
+  AclRevisionInformation = 1, 
   AclSizeInformation
 } ACL_INFORMATION_CLASS;
 
@@ -4947,12 +4910,12 @@ typedef struct _LUID {
     LONG HighPart;
 } LUID, *PLUID;
 
-#pragma pack(push,4)
+#include <pshpack4.h>
 typedef struct _LUID_AND_ATTRIBUTES {
   LUID   Luid;
   DWORD  Attributes;
 } LUID_AND_ATTRIBUTES, *PLUID_AND_ATTRIBUTES;
-#pragma pack(pop)
+#include <poppack.h>
 
 /*
  * PRIVILEGE_SET
@@ -5049,7 +5012,7 @@ typedef struct _SECURITY_QUALITY_OF_SERVICE {
  * TOKEN_STATISTICS
  */
 
-#pragma pack(push,4)
+#include <pshpack4.h>
 typedef struct _TOKEN_STATISTICS {
   LUID  TokenId;
   LUID  AuthenticationId;
@@ -5062,7 +5025,7 @@ typedef struct _TOKEN_STATISTICS {
   DWORD PrivilegeCount;
   LUID  ModifiedId;
 } TOKEN_STATISTICS;
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef struct _TOKEN_GROUPS_AND_PRIVILEGES {
   DWORD                 SidCount;
@@ -6223,14 +6186,6 @@ typedef struct _OBJECT_TYPE_LIST {
     GUID *ObjectType;
 } OBJECT_TYPE_LIST, *POBJECT_TYPE_LIST;
 
-typedef enum _AUDIT_EVENT_TYPE
-{
-    AuditEventObjectAccess,
-    AuditEventDirectoryServiceAccess
-} AUDIT_EVENT_TYPE, *PAUDIT_EVENT_TYPE;
-
-#define AUDIT_ALLOW_NO_PRIVILEGE 0x1
-
 typedef struct _RTL_CRITICAL_SECTION_DEBUG
 {
   WORD   Type;
@@ -6303,16 +6258,16 @@ typedef struct _RTL_BARRIER {
     DWORD Reserved5;
 } RTL_BARRIER, *PRTL_BARRIER;
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct _IO_COUNTERS {
-    ULONGLONG ReadOperationCount;
-    ULONGLONG WriteOperationCount;
-    ULONGLONG OtherOperationCount;
-    ULONGLONG ReadTransferCount;
-    ULONGLONG WriteTransferCount;
-    ULONGLONG OtherTransferCount;
+    ULONGLONG DECLSPEC_ALIGN(8) ReadOperationCount;
+    ULONGLONG DECLSPEC_ALIGN(8) WriteOperationCount;
+    ULONGLONG DECLSPEC_ALIGN(8) OtherOperationCount;
+    ULONGLONG DECLSPEC_ALIGN(8) ReadTransferCount;
+    ULONGLONG DECLSPEC_ALIGN(8) WriteTransferCount;
+    ULONGLONG DECLSPEC_ALIGN(8) OtherTransferCount;
 } IO_COUNTERS, *PIO_COUNTERS;
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef struct {
 	DWORD dwOSVersionInfoSize;
@@ -6558,57 +6513,15 @@ typedef struct _ACTIVATION_CONTEXT_DATA_DLL_REDIRECTION_PATH_SEGMENT
 typedef enum _JOBOBJECTINFOCLASS
 {
     JobObjectBasicAccountingInformation = 1,
-    JobObjectBasicLimitInformation = 2,
-    JobObjectBasicProcessIdList = 3,
-    JobObjectBasicUIRestrictions = 4,
-    JobObjectSecurityLimitInformation = 5,
-    JobObjectEndOfJobTimeInformation = 6,
-    JobObjectAssociateCompletionPortInformation = 7,
-    JobObjectBasicAndIoAccountingInformation = 8,
-    JobObjectExtendedLimitInformation = 9,
-    JobObjectJobSetInformation = 10,
-    JobObjectGroupInformation = 11,
-    JobObjectNotificationLimitInformation = 12,
-    JobObjectLimitViolationInformation = 13,
-    JobObjectGroupInformationEx = 14,
-    JobObjectCpuRateControlInformation = 15,
-    JobObjectCompletionFilter = 16,
-    JobObjectCompletionCounter = 17,
-    JobObjectFreezeInformation = 18,
-    JobObjectExtendedAccountingInformation = 19,
-    JobObjectWakeInformation = 20,
-    JobObjectBackgroundInformation = 21,
-    JobObjectSchedulingRankBiasInformation = 22,
-    JobObjectTimerVirtualizationInformation = 23,
-    JobObjectCycleTimeNotification = 24,
-    JobObjectClearEvent = 25,
-    JobObjectInterferenceInformation = 26,
-    JobObjectClearPeakJobMemoryUsed = 27,
-    JobObjectMemoryUsageInformation = 28,
-    JobObjectSharedCommit = 29,
-    JobObjectContainerId = 30,
-    JobObjectIoRateControlInformation = 31,
-    JobObjectNetRateControlInformation = 32,
-    JobObjectNotificationLimitInformation2 = 33,
-    JobObjectLimitViolationInformation2 = 34,
-    JobObjectCreateSilo = 35,
-    JobObjectSiloBasicInformation = 36,
-    JobObjectSiloRootDirectory = 37,
-    JobObjectServerSiloBasicInformation = 38,
-    JobObjectServerSiloUserSharedData = 39,
-    JobObjectServerSiloInitialize = 40,
-    JobObjectServerSiloRunningState = 41,
-    JobObjectIoAttribution = 42,
-    JobObjectMemoryPartitionInformation = 43,
-    JobObjectContainerTelemetryId = 44,
-    JobObjectSiloSystemRoot = 45,
-    JobObjectEnergyTrackingState = 46,
-    JobObjectThreadImpersonationInformation = 47,
-    JobObjectIoPriorityLimit = 48,
-    JobObjectPagePriorityLimit = 49,
-    JobObjectServerSiloDiagnosticInformation = 50,
-    JobObjectNetworkAccountingInformation = 51,
-    JobObjectCpuPartition = 52,
+    JobObjectBasicLimitInformation,
+    JobObjectBasicProcessIdList,
+    JobObjectBasicUIRestrictions,
+    JobObjectSecurityLimitInformation,
+    JobObjectEndOfJobTimeInformation,
+    JobObjectAssociateCompletionPortInformation,
+    JobObjectBasicAndIoAccountingInformation,
+    JobObjectExtendedLimitInformation,
+    JobObjectJobSetInformation,
     MaxJobObjectInfoClass
 } JOBOBJECTINFOCLASS;
 
@@ -6774,13 +6687,8 @@ typedef struct _PROCESSOR_RELATIONSHIP
 typedef struct _NUMA_NODE_RELATIONSHIP
 {
     DWORD NodeNumber;
-    BYTE Reserved[18];
-    WORD GroupCount;
-    union
-    {
-        GROUP_AFFINITY GroupMask;
-        GROUP_AFFINITY GroupMasks[ANYSIZE_ARRAY];
-    };
+    BYTE Reserved[20];
+    GROUP_AFFINITY GroupMask;
 } NUMA_NODE_RELATIONSHIP, *PNUMA_NODE_RELATIONSHIP;
 
 typedef struct _CACHE_RELATIONSHIP
@@ -6790,13 +6698,8 @@ typedef struct _CACHE_RELATIONSHIP
     WORD LineSize;
     DWORD CacheSize;
     PROCESSOR_CACHE_TYPE Type;
-    BYTE Reserved[18];
-    WORD GroupCount;
-    union
-    {
-        GROUP_AFFINITY GroupMask;
-        GROUP_AFFINITY GroupMasks[ANYSIZE_ARRAY];
-    };
+    BYTE Reserved[20];
+    GROUP_AFFINITY GroupMask;
 } CACHE_RELATIONSHIP, *PCACHE_RELATIONSHIP;
 
 typedef struct _GROUP_RELATIONSHIP
@@ -7179,7 +7082,7 @@ static FORCEINLINE __int64 InterlockedAdd64( __int64 volatile *dest, __int64 val
 
 static FORCEINLINE void MemoryBarrier(void)
 {
-    long volatile dummy = 0;
+    LONG dummy;
     InterlockedOr(&dummy, 0);
 }
 
@@ -7217,16 +7120,13 @@ static FORCEINLINE void MemoryBarrier(void)
 #pragma intrinsic(__iso_volatile_load32)
 #pragma intrinsic(__iso_volatile_load64)
 #pragma intrinsic(__iso_volatile_store32)
-#pragma intrinsic(__iso_volatile_store64)
 #define __WINE_LOAD32_NO_FENCE(src) (__iso_volatile_load32(src))
 #define __WINE_LOAD64_NO_FENCE(src) (__iso_volatile_load64(src))
 #define __WINE_STORE32_NO_FENCE(dest, value) (__iso_volatile_store32(dest, value))
-#define __WINE_STORE64_NO_FENCE(dest, value) (__iso_volatile_store64(dest, value))
 #else  /* _MSC_VER >= 1700 */
 #define __WINE_LOAD32_NO_FENCE(src) (*(src))
 #define __WINE_LOAD64_NO_FENCE(src) (*(src))
 #define __WINE_STORE32_NO_FENCE(dest, value) ((void)(*(dest) = (value)))
-#define __WINE_STORE64_NO_FENCE(dest, value) ((void)(*(dest) = (value)))
 #endif  /* _MSC_VER >= 1700 */
 
 #if defined(__i386__) || defined(__x86_64__)
@@ -7253,22 +7153,6 @@ static FORCEINLINE LONG ReadAcquire( LONG const volatile *src )
     return value;
 }
 
-static FORCEINLINE LONG64 ReadAcquire64( LONG64 const volatile *src )
-{
-    LONG64 value;
-#if defined(__i386__) && _MSC_VER < 1700
-    __asm {
-        mov   eax, src
-        fild  qword ptr [eax]
-        fistp value
-    }
-#else
-    value = __WINE_LOAD64_NO_FENCE( (__int64 const volatile *)src );
-    __wine_memory_barrier_acq_rel();
-#endif
-    return value;
-}
-
 static FORCEINLINE LONG ReadNoFence( LONG const volatile *src )
 {
     LONG value = __WINE_LOAD32_NO_FENCE( (int const volatile *)src );
@@ -7277,16 +7161,7 @@ static FORCEINLINE LONG ReadNoFence( LONG const volatile *src )
 
 static FORCEINLINE LONG64 ReadNoFence64( LONG64 const volatile *src )
 {
-    LONG64 value;
-#if defined(__i386__) && _MSC_VER < 1700
-    __asm {
-        mov   eax, src
-        fild  qword ptr [eax]
-        fistp value
-    }
-#else
-    value = __WINE_LOAD64_NO_FENCE( (__int64 const volatile *)src );
-#endif
+    LONG64 value = __WINE_LOAD64_NO_FENCE( (__int64 const volatile *)src );
     return value;
 }
 
@@ -7296,30 +7171,11 @@ static FORCEINLINE void WriteRelease( LONG volatile *dest, LONG value )
     __WINE_STORE32_NO_FENCE( (int volatile *)dest, value );
 }
 
-static FORCEINLINE void WriteRelease64( LONG64 volatile *dest, LONG64 value )
-{
-#if defined(__i386__) && _MSC_VER < 1700
-    __asm {
-        mov   eax, dest
-        fild  value
-        fistp qword ptr [eax]
-    }
-#else
-    __wine_memory_barrier_acq_rel();
-    __WINE_STORE64_NO_FENCE( (__int64 volatile *)dest, value );
-#endif
-}
-
 static FORCEINLINE void WriteNoFence( LONG volatile *dest, LONG value )
 {
     __WINE_STORE32_NO_FENCE( (int volatile *)dest, value );
 }
 
-static FORCEINLINE void WriteNoFence64( LONG64 volatile *dest, LONG64 value )
-{
-    __WINE_STORE64_NO_FENCE( (__int64 volatile *)dest, value );
-}
-
 #elif defined(__GNUC__)
 
 static FORCEINLINE BOOLEAN WINAPI BitScanForward(DWORD *index, DWORD mask)
@@ -7470,10 +7326,10 @@ static FORCEINLINE void MemoryBarrier(void)
 
 #if defined(__x86_64__) || defined(__i386__)
 /* On x86, Support old GCC with either no or buggy (GCC BZ#81316) __atomic_* support */
-#define __WINE_ATOMIC_LOAD_ACQUIRE(ptr, ret) do { C_ASSERT(sizeof(*(ptr)) <= sizeof(void *)); *(ret) = *(ptr); __asm__ __volatile__( "" ::: "memory" ); } while (0)
-#define __WINE_ATOMIC_LOAD_RELAXED(ptr, ret) do { C_ASSERT(sizeof(*(ptr)) <= sizeof(void *)); *(ret) = *(ptr); } while (0)
-#define __WINE_ATOMIC_STORE_RELEASE(ptr, val) do { C_ASSERT(sizeof(*(ptr)) <= sizeof(void *)); __asm__ __volatile__( "" ::: "memory" ); *(ptr) = *(val); } while (0)
-#define __WINE_ATOMIC_STORE_RELAXED(ptr, val) do { C_ASSERT(sizeof(*(ptr)) <= sizeof(void *)); *(ptr) = *(val); } while (0)
+#define __WINE_ATOMIC_LOAD_ACQUIRE(ptr, ret) do { *(ret) = *(ptr); __asm__ __volatile__( "" ::: "memory" ); } while (0)
+#define __WINE_ATOMIC_LOAD_RELAXED(ptr, ret) do { *(ret) = *(ptr); } while (0)
+#define __WINE_ATOMIC_STORE_RELEASE(ptr, val) do { __asm__ __volatile__( "" ::: "memory" ); *(ptr) = *(val); } while (0)
+#define __WINE_ATOMIC_STORE_RELAXED(ptr, val) do { *(ptr) = *(val); } while (0)
 #else
 #define __WINE_ATOMIC_LOAD_ACQUIRE(ptr, ret) __atomic_load(ptr, ret, __ATOMIC_ACQUIRE)
 #define __WINE_ATOMIC_LOAD_RELAXED(ptr, ret) __atomic_load(ptr, ret, __ATOMIC_RELAXED)
@@ -7488,17 +7344,6 @@ static FORCEINLINE LONG ReadAcquire( LONG const volatile *src )
     return value;
 }
 
-static FORCEINLINE LONG64 ReadAcquire64( LONG64 const volatile *src )
-{
-    LONG64 value;
-#ifdef __i386__
-    __asm__ __volatile__( "fildq %1\n\tfistpq %0" : "=m" (value) : "m" (*src) : "memory", "st" );
-#else
-    __WINE_ATOMIC_LOAD_ACQUIRE( src, &value );
-#endif
-    return value;
-}
-
 static FORCEINLINE LONG ReadNoFence( LONG const volatile *src )
 {
     LONG value;
@@ -7509,11 +7354,7 @@ static FORCEINLINE LONG ReadNoFence( LONG const volatile *src )
 static FORCEINLINE LONG64 ReadNoFence64( LONG64 const volatile *src )
 {
     LONG64 value;
-#ifdef __i386__
-    __asm__ __volatile__( "fildq %1\n\tfistpq %0" : "=m" (value) : "m" (*src) : "memory", "st" );
-#else
     __WINE_ATOMIC_LOAD_RELAXED( src, &value );
-#endif
     return value;
 }
 
@@ -7522,29 +7363,11 @@ static FORCEINLINE void WriteRelease( LONG volatile *dest, LONG value )
     __WINE_ATOMIC_STORE_RELEASE( dest, &value );
 }
 
-static FORCEINLINE void WriteRelease64( LONG64 volatile *dest, LONG64 value )
-{
-#ifdef __i386__
-    __asm__ __volatile__( "fildq %1\n\tfistpq %0" : "=m" (*dest) : "m" (value) : "memory", "st" );
-#else
-    __WINE_ATOMIC_STORE_RELEASE( dest, &value );
-#endif
-}
-
 static FORCEINLINE void WriteNoFence( LONG volatile *dest, LONG value )
 {
     __WINE_ATOMIC_STORE_RELAXED( dest, &value );
 }
 
-static FORCEINLINE void WriteNoFence64( LONG64 volatile *dest, LONG64 value )
-{
-#ifdef __i386__
-    __asm__ __volatile__( "fildq %1\n\tfistpq %0" : "=m" (*dest) : "m" (value) : "memory", "st" );
-#else
-    __WINE_ATOMIC_STORE_RELAXED( dest, &value );
-#endif
-}
-
 static FORCEINLINE DECLSPEC_NORETURN void __fastfail(unsigned int code)
 {
 #if defined(__x86_64__) || defined(__i386__)
@@ -7591,68 +7414,12 @@ static FORCEINLINE unsigned char InterlockedCompareExchange128( volatile __int64
 #define InterlockedDecrementSizeT(a) InterlockedDecrement64((LONGLONG *)(a))
 #define InterlockedExchangeAddSizeT(a, b) InterlockedExchangeAdd64((LONGLONG *)(a), (b))
 #define InterlockedIncrementSizeT(a) InterlockedIncrement64((LONGLONG *)(a))
-#define ReadLongPtrAcquire   ReadAcquire64
-#define ReadLongPtrNoFence   ReadNoFence64
-#define ReadULongPtrAcquire  ReadULong64Acquire
-#define ReadULongPtrNoFence  ReadULong64NoFence
-#define WriteLongPtrRelease  WriteRelease64
-#define WriteLongPtrNoFence  WriteNoFence64
-#define WriteULongPtrRelease WriteULong64Release
-#define WriteULongPtrNoFence WriteULong64NoFence
-
-static FORCEINLINE void *ReadPointerAcquire( void* const volatile *src )
-{
-    return (void *)ReadAcquire64( (const volatile LONG64 *)src );
-}
-
-static FORCEINLINE void *ReadPointerNoFence( void* const volatile *src )
-{
-    return (void *)ReadNoFence64( (const volatile LONG64 *)src );
-}
-
-static FORCEINLINE void WritePointerRelease( void* volatile *dest, void* value )
-{
-    WriteRelease64( (volatile LONG64 *)dest, (LONG64)value );
-}
-
-static FORCEINLINE void WritePointerNoFence( void* volatile *dest, void* value )
-{
-    WriteNoFence64( (volatile LONG64 *)dest, (LONG64)value );
-}
 
 #else /* _WIN64 */
 
 #define InterlockedDecrementSizeT(a) InterlockedDecrement((LONG *)(a))
 #define InterlockedExchangeAddSizeT(a, b) InterlockedExchangeAdd((LONG *)(a), (b))
 #define InterlockedIncrementSizeT(a) InterlockedIncrement((LONG *)(a))
-#define ReadLongPtrAcquire   ReadAcquire
-#define ReadLongPtrNoFence   ReadNoFence
-#define ReadULongPtrAcquire  ReadULongAcquire
-#define ReadULongPtrNoFence  ReadULongNoFence
-#define WriteLongPtrRelease  WriteRelease
-#define WriteLongPtrNoFence  WriteNoFence
-#define WriteULongPtrRelease WriteULongRelease
-#define WriteULongPtrNoFence WriteULongNoFence
-
-static FORCEINLINE void *ReadPointerAcquire( void* const volatile *src )
-{
-    return (void *)ReadAcquire( (const volatile LONG *)src );
-}
-
-static FORCEINLINE void *ReadPointerNoFence( void* const volatile *src )
-{
-    return (void *)ReadNoFence( (const volatile LONG *)src );
-}
-
-static FORCEINLINE void WritePointerRelease( void* volatile *dest, void* value )
-{
-    WriteRelease( (volatile LONG *)dest, (LONG)value );
-}
-
-static FORCEINLINE void WritePointerNoFence( void* volatile *dest, void* value )
-{
-    WriteNoFence( (volatile LONG *)dest, (LONG)value );
-}
 
 #endif /* _WIN64 */
 
@@ -7669,36 +7436,6 @@ static FORCEINLINE void YieldProcessor(void)
 #endif
 }
 
-#if defined(__x86_64__)
-# if defined(__arm64ec__)
-#  define __shiftright128 ShiftRight128
-#  define _umul128 UnsignedMultiply128
-# else
-#  define ShiftRight128 __shiftright128
-#  define UnsignedMultiply128 _umul128
-#  if defined(_MSC_VER)
-DWORD64 __shiftright128(DWORD64,DWORD64,BYTE);
-DWORD64 _umul128(DWORD64,DWORD64,DWORD64*);
-#   pragma intrinsic(__shiftright128)
-#   pragma intrinsic(_umul128)
-#  endif
-# endif
-#endif
-
-#if (defined(__x86_64__) && !defined(_MSC_VER)) || defined(__aarch64__) || defined(__arm64ec__)
-static FORCEINLINE DWORD64 ShiftRight128( DWORD64 lo, DWORD64 hi, BYTE shift )
-{
-    return ((unsigned __int128)hi << 64 | lo) >> shift;
-}
-
-static FORCEINLINE DWORD64 UnsignedMultiply128( DWORD64 a, DWORD64 b, DWORD64 *hi )
-{
-    unsigned __int128 v = (unsigned __int128)a * b;
-    *hi = v >> 64;
-    return (DWORD64)v;
-}
-#endif
-
 #ifdef __cplusplus
 }
 #endif
