diff --git a/include/winperf.h b/include/winperf.h
index 3b808aff49852bb25ab67ee5353de72f025f3cb6..926da8f8c4a84f7c0d7a8807a0e115ff21505801 100644
--- a/include/winperf.h
+++ b/include/winperf.h
@@ -67,7 +67,7 @@
 #define PERF_DETAIL_EXPERT      300
 #define PERF_DETAIL_WIZARD      400
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 
 /* Performance data structure header
  * returned in answer to HKEY_PERFORMANCE_DATA request
@@ -191,7 +191,7 @@ typedef struct _PERF_COUNTER_BLOCK
 #define PERF_COUNTER_COUNTER (PERF_SIZE_DWORD | PERF_TYPE_COUNTER | PERF_COUNTER_RATE | \
                               PERF_TIMER_TICK | PERF_DELTA_COUNTER | PERF_DISPLAY_PER_SEC)
 
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef DWORD (APIENTRY PM_OPEN_PROC)(LPWSTR);
 typedef DWORD (APIENTRY PM_COLLECT_PROC)(LPWSTR,LPVOID *,LPDWORD,LPDWORD);
