diff --git a/include/winsafer.h b/include/winsafer.h
index 38ca57c8a1cd662d8c9245b42ba56708fec1cf9b..02150c599e7144fb13189809e041326ece4d456a 100644
--- a/include/winsafer.h
+++ b/include/winsafer.h
@@ -37,7 +37,7 @@ DECLARE_HANDLE(SAFER_LEVEL_HANDLE);
 
 #define SAFER_MAX_HASH_SIZE 64
 
-#pragma pack(push,8)
+#include <pshpack8.h>
 typedef struct _SAFER_CODE_PROPERTIES_V2 {
     DWORD cbSize;
     DWORD dwCheckFlags;
@@ -59,7 +59,7 @@ typedef struct _SAFER_CODE_PROPERTIES_V2 {
 } SAFER_CODE_PROPERTIES_V2, *PSAFER_CODE_PROPERTIES_V2;
 
 typedef SAFER_CODE_PROPERTIES_V2 SAFER_CODE_PROPERTIES, *PSAFER_CODE_PROPERTIES;
-#pragma pack(pop)
+#include <poppack.h>
 
 #define SAFER_SCOPEID_MACHINE  1
 #define SAFER_SCOPEID_USER     2
