diff --git a/include/wmsdkidl.idl b/include/wmsdkidl.idl
index 48bcdc713f19c631f4e1483e1edbfc21b1a7d9b2..55142f47e93e08244c93e94181c278fc95ef3a38 100644
--- a/include/wmsdkidl.idl
+++ b/include/wmsdkidl.idl
@@ -210,7 +210,7 @@ typedef enum tagWMT_STORAGE_FORMAT
     WMT_Storage_Format_V1   = 1
 } WMT_STORAGE_FORMAT;
 
-#pragma pack(push,2)
+#include <pshpack2.h>
 typedef struct _WMStreamPrioritizationRecord
 {
     WORD wStreamNumber;
@@ -224,7 +224,7 @@ typedef struct _WMT_TIMECODE_EXTENSION_DATA
     DWORD dwUserbits;
     DWORD dwAmFlags;
 } WMT_TIMECODE_EXTENSION_DATA;
-#pragma pack(pop)
+#include <poppack.h>
 
 typedef struct _WM_PORT_NUMBER_RANGE
 {
