diff --git a/include/ws2def.h b/include/ws2def.h
index dc3201924c14754013270a4d7b883190ba3ba9b6..46fb80db470561160cae41861191b5053fcbc95b 100644
--- a/include/ws2def.h
+++ b/include/ws2def.h
@@ -65,14 +65,7 @@ typedef struct WS(sockaddr)
 #define AF_CLUSTER                 24
 #define AF_12844                   25
 #define AF_IRDA                    26
-#define AF_NETDES                  28
-#define AF_TCNPROCESS              29
-#define AF_TCNMESSAGE              30
-#define AF_ICLFXBM                 31
-#define AF_BTH                     32
-#define AF_LINK                    33
-#define AF_HYPERV                  34
-#define AF_MAX                     35
+#define AF_MAX                     27
 #else /* USE_WS_PREFIX */
 #define WS_AF_UNSPEC               0
 #define WS_AF_UNIX                 1
@@ -103,14 +96,7 @@ typedef struct WS(sockaddr)
 #define WS_AF_CLUSTER              24
 #define WS_AF_12844                25
 #define WS_AF_IRDA                 26
-#define WS_AF_NETDES               28
-#define WS_AF_TCNPROCESS           29
-#define WS_AF_TCNMESSAGE           30
-#define WS_AF_ICLFXBM              31
-#define WS_AF_BTH                  32
-#define WS_AF_LINK                 33
-#define WS_AF_HYPERV               34
-#define WS_AF_MAX                  35
+#define WS_AF_MAX                  27
 #endif /* USE_WS_PREFIX */
 
 #ifndef USE_WS_PREFIX
