diff --git a/include/ws2spi.h b/include/ws2spi.h
index f2d21068a4f27d78727771edc5de131f3b966c7d..7e214a091b2065ce7197b4597c8b88e5a09a8292 100644
--- a/include/ws2spi.h
+++ b/include/ws2spi.h
@@ -25,7 +25,7 @@
 #include <winsock2.h>
 #endif /* !defined(_WINSOCK2API_) */
 
-#pragma pack(push,4)
+#include <pshpack4.h>
 
 #ifdef __cplusplus
 extern "C" {
@@ -58,6 +58,6 @@ INT      WINAPI WSCWriteProviderOrder(LPDWORD,DWORD);
 } /* extern "C" */
 #endif /* defined(__cplusplus) */
 
-#pragma pack(pop)
+#include <poppack.h>
 
 #endif /* !defined(_WINSOCK2SPI_) */
