diff --git a/libs/compiler-rt/lib/builtins/divmoddi4.c b/libs/compiler-rt/lib/builtins/divmoddi4.c
deleted file mode 100644
index 0d4df67a63e0b57bf24fa8c4eddb0192c31e58d1..0000000000000000000000000000000000000000
--- a/libs/compiler-rt/lib/builtins/divmoddi4.c
+++ /dev/null
@@ -1,25 +0,0 @@
-/*===-- divmoddi4.c - Implement __divmoddi4 --------------------------------===
- *
- *                    The LLVM Compiler Infrastructure
- *
- * This file is dual licensed under the MIT and the University of Illinois Open
- * Source Licenses. See LICENSE.TXT for details.
- *
- * ===----------------------------------------------------------------------===
- *
- * This file implements __divmoddi4 for the compiler_rt library.
- *
- * ===----------------------------------------------------------------------===
- */
-
-#include "int_lib.h"
-
-/* Returns: a / b, *rem = a % b  */
-
-COMPILER_RT_ABI di_int
-__divmoddi4(di_int a, di_int b, di_int* rem)
-{
-  di_int d = __divdi3(a,b);
-  *rem = a - (d*b);
-  return d;
-}
