diff --git a/libs/faudio/include/FAudioFX.h b/libs/faudio/include/FAudioFX.h
index 27bba87893d997bca6b8b226b9c8013080fb8513..2ed0d2ab8f72774b217937a18c903db5d33acd4f 100644
--- a/libs/faudio/include/FAudioFX.h
+++ b/libs/faudio/include/FAudioFX.h
@@ -56,11 +56,6 @@ typedef struct FAudioFXVolumeMeterLevels
 	uint32_t ChannelCount;
 } FAudioFXVolumeMeterLevels;
 
-typedef struct FAudioFXCollectorState
-{
-	uint32_t WriteOffset;
-} FAudioFXCollectorState;
-
 typedef struct FAudioFXReverbParameters
 {
 	float WetDryMix;
@@ -267,21 +262,11 @@ typedef struct FAudioFXReverbI3DL2Parameters
 
 /* Functions */
 
-FAUDIOAPI uint32_t FAudioCreateCollectorEXT(FAPO** ppApo, uint32_t Flags, float* pBuffer, uint32_t bufferLength);
 FAUDIOAPI uint32_t FAudioCreateVolumeMeter(FAPO** ppApo, uint32_t Flags);
 FAUDIOAPI uint32_t FAudioCreateReverb(FAPO** ppApo, uint32_t Flags);
 FAUDIOAPI uint32_t FAudioCreateReverb9(FAPO** ppApo, uint32_t Flags);
 
 /* See "extensions/CustomAllocatorEXT.txt" for more information. */
-FAUDIOAPI uint32_t FAudioCreateCollectorWithCustomAllocatorEXT(
-	FAPO** ppApo,
-	uint32_t Flags,
-	float* pBuffer,
-	uint32_t bufferLength,
-	FAudioMallocFunc customMalloc,
-	FAudioFreeFunc customFree,
-	FAudioReallocFunc customRealloc
-);
 FAUDIOAPI uint32_t FAudioCreateVolumeMeterWithCustomAllocatorEXT(
 	FAPO** ppApo,
 	uint32_t Flags,
