diff --git a/libs/faudio/src/FAudio_internal.h b/libs/faudio/src/FAudio_internal.h
index 7fb99666ba546167e8505d6b3adb6ddbd10c1b1d..9cbc12cb4689023487844b108913e919077192e4 100644
--- a/libs/faudio/src/FAudio_internal.h
+++ b/libs/faudio/src/FAudio_internal.h
@@ -27,7 +27,6 @@
 #include "FAudio.h"
 #include "FAPOBase.h"
 #include <stdarg.h>
-#include <stdbool.h>
 
 
 #ifdef FAUDIO_WIN32_PLATFORM
@@ -42,6 +41,9 @@
 #define WIN32_LEAN_AND_MEAN
 #include <windows.h>
 
+#define NO_MEMCPY_OVERRIDE
+#define NO_MEMSET_OVERRIDE
+
 #define FAudio_malloc malloc
 #define FAudio_realloc realloc
 #define FAudio_free free
@@ -126,13 +128,6 @@ extern void FAudio_Log(char const *msg);
 #define FAudio_swap64LE(x) SDL_Swap64LE(x)
 #define FAudio_swap64BE(x) SDL_Swap64BE(x)
 
-/* SDL3 allows memcpy/memset for compiler optimization reasons */
-#ifdef SDL_SLOW_MEMCPY
-#define STB_MEMCPY_OVERRIDE
-#endif
-#ifdef SDL_SLOW_MEMSET
-#define STB_MEMSET_OVERRIDE
-#endif
 #else
 #include <SDL_stdinc.h>
 #include <SDL_assert.h>
@@ -145,9 +140,14 @@ extern void FAudio_Log(char const *msg);
 #define FAudio_swap32BE(x) SDL_SwapBE32(x)
 #define FAudio_swap64LE(x) SDL_SwapLE64(x)
 #define FAudio_swap64BE(x) SDL_SwapBE64(x)
+#endif
 
-#define STB_MEMCPY_OVERRIDE
-#define STB_MEMSET_OVERRIDE
+/* SDL3 allows memcpy/memset for compiler optimization reasons */
+#ifndef SDL_SLOW_MEMCPY
+#define NO_MEMCPY_OVERRIDE
+#endif
+#ifndef SDL_SLOW_MEMSET
+#define NO_MEMSET_OVERRIDE
 #endif
 
 #define FAudio_malloc SDL_malloc
