diff --git a/libs/fluidsynth/glib.c b/libs/fluidsynth/glib.c
index 5756583111a332aa55f8f92be413ceaab2980932..cd938a4fb6bb4659db626de3fa36f1cb956efd03 100644
--- a/libs/fluidsynth/glib.c
+++ b/libs/fluidsynth/glib.c
@@ -40,9 +40,10 @@ int g_snprintf( char *buffer, size_t size, const char *format, ... )
 
 double g_get_monotonic_time(void)
 {
-    LARGE_INTEGER counter, frequency;
+    static LARGE_INTEGER frequency = {0};
+    LARGE_INTEGER counter;
 
-    QueryPerformanceFrequency( &frequency );
+    if (!frequency.QuadPart) QueryPerformanceFrequency( &frequency );
     QueryPerformanceCounter( &counter );
 
     return counter.QuadPart * 1000000.0 / frequency.QuadPart; /* time in micros */
@@ -50,15 +51,15 @@ double g_get_monotonic_time(void)
 
 void g_usleep( unsigned int micros )
 {
-    Sleep( (micros + 999) / 1000 );
+    Sleep( micros / 1000 );
 }
 
 static DWORD CALLBACK g_thread_wrapper( void *args )
 {
     GThread *thread = args;
-    thread->result = thread->func( thread->data );
-    g_thread_unref( thread );
-    return 0;
+    gpointer ret = thread->func( thread->data );
+    if (!InterlockedDecrement( &thread->ref )) free( thread );
+    return (UINT_PTR)ret;
 }
 
 GThread *g_thread_try_new( const char *name, GThreadFunc func, gpointer data, GError **err )
@@ -81,21 +82,14 @@ GThread *g_thread_try_new( const char *name, GThreadFunc func, gpointer data, GE
 
 void g_thread_unref( GThread *thread )
 {
-    if (!InterlockedDecrement( &thread->ref ))
-    {
-        CloseHandle( thread->handle );
-        free( thread );
-    }
+    CloseHandle( thread->handle );
+    if (!InterlockedDecrement( &thread->ref )) free( thread );
 }
 
-gpointer g_thread_join( GThread *thread )
+void g_thread_join( GThread *thread )
 {
-    gpointer result;
-
     WaitForSingleObject( thread->handle, INFINITE );
-    result = thread->result;
     g_thread_unref( thread );
-    return result;
 }
 
 void g_clear_error( GError **error )
@@ -106,10 +100,7 @@ void g_clear_error( GError **error )
 int g_file_test( const char *path, int test )
 {
     DWORD attrs = GetFileAttributesA( path );
-    if (attrs != INVALID_FILE_ATTRIBUTES)
-    {
-        if (test & G_FILE_TEST_EXISTS) return 1;
-        if ((test & G_FILE_TEST_IS_REGULAR) && !(attrs & (FILE_ATTRIBUTE_DIRECTORY | FILE_ATTRIBUTE_DEVICE))) return 1;
-    }
+    if (test == G_FILE_TEST_EXISTS) return attrs != INVALID_FILE_ATTRIBUTES;
+    if (test == G_FILE_TEST_IS_REGULAR) return attrs == FILE_ATTRIBUTE_NORMAL;
     return 0;
 }
