diff --git a/libs/fluidsynth/include/fluidsynth/version.h b/libs/fluidsynth/include/fluidsynth/version.h
index f15b9b2b680b488894dcc4349a2f76a93c695abf..b1dcf349ba61c80e6554f57494eeb577e9d55c14 100644
--- a/libs/fluidsynth/include/fluidsynth/version.h
+++ b/libs/fluidsynth/include/fluidsynth/version.h
@@ -31,10 +31,10 @@ extern "C" {
  *
  * @{
  */
-#define FLUIDSYNTH_VERSION       "2.4.2"         /**< String constant of libfluidsynth version. */
+#define FLUIDSYNTH_VERSION       "2.3.6"           /**< String constant of libfluidsynth version. */
 #define FLUIDSYNTH_VERSION_MAJOR 2     /**< libfluidsynth major version integer constant. */
-#define FLUIDSYNTH_VERSION_MINOR 4     /**< libfluidsynth minor version integer constant. */
-#define FLUIDSYNTH_VERSION_MICRO 2     /**< libfluidsynth micro version integer constant. */
+#define FLUIDSYNTH_VERSION_MINOR 3     /**< libfluidsynth minor version integer constant. */
+#define FLUIDSYNTH_VERSION_MICRO 6     /**< libfluidsynth micro version integer constant. */
 
 FLUIDSYNTH_API void fluid_version(int *major, int *minor, int *micro);
 FLUIDSYNTH_API char* fluid_version_str(void);
