diff --git a/libs/fluidsynth/src/rvoice/fluid_iir_filter.c b/libs/fluidsynth/src/rvoice/fluid_iir_filter.c
index ba8d5a9603fa25ba55ad0314b18f88eae707b10d..79d293c61b2b8b12ba25bde5cb38c3ff483c66fe 100644
--- a/libs/fluidsynth/src/rvoice/fluid_iir_filter.c
+++ b/libs/fluidsynth/src/rvoice/fluid_iir_filter.c
@@ -22,13 +22,6 @@
 #include "fluid_sys.h"
 #include "fluid_conv.h"
 
-
-static FLUID_INLINE void
-fluid_iir_filter_calculate_coefficients(fluid_iir_filter_t *iir_filter, fluid_real_t output_rate,
-                                        fluid_real_t *a1_out, fluid_real_t *a2_out,
-                                        fluid_real_t *b02_out, fluid_real_t *b1_out);
-
-
 /**
  * Applies a low- or high-pass filter with variable cutoff frequency and quality factor
  * for a given biquad transfer function:
@@ -55,9 +48,9 @@ fluid_iir_filter_calculate_coefficients(fluid_iir_filter_t *iir_filter, fluid_re
  */
 void
 fluid_iir_filter_apply(fluid_iir_filter_t *iir_filter,
-                       fluid_real_t *dsp_buf, int count, fluid_real_t output_rate)
+                       fluid_real_t *dsp_buf, int count)
 {
-    if(iir_filter->type == FLUID_IIR_DISABLED || FLUID_FABS(iir_filter->last_q) <= 0.001)
+    if(iir_filter->type == FLUID_IIR_DISABLED || iir_filter->q_lin == 0)
     {
         return;
     }
@@ -73,9 +66,6 @@ fluid_iir_filter_apply(fluid_iir_filter_t *iir_filter,
         fluid_real_t dsp_b02 = iir_filter->b02;
         fluid_real_t dsp_b1 = iir_filter->b1;
 
-        int fres_incr_count = iir_filter->fres_incr_count;
-        int q_incr_count = iir_filter->q_incr_count;
-
         fluid_real_t dsp_centernode;
         int dsp_i;
 
@@ -104,24 +94,6 @@ fluid_iir_filter_apply(fluid_iir_filter_t *iir_filter,
             // dsp_buf[dsp_i] = dsp_b02 * dsp_input + dsp_hist1;
             // dsp_hist1 = dsp_b1 * dsp_input - dsp_a1 * dsp_buf[dsp_i] + dsp_hist2;
             // dsp_hist2 = dsp_b02 * dsp_input - dsp_a2 * dsp_buf[dsp_i];
-
-            if(fres_incr_count > 0 || q_incr_count > 0)
-            {
-                if(fres_incr_count > 0)
-                {
-                    --fres_incr_count;
-                    iir_filter->last_fres += iir_filter->fres_incr;
-                }
-                if(q_incr_count > 0)
-                {
-                    --q_incr_count;
-                    iir_filter->last_q += iir_filter->q_incr;
-                }
-
-                LOG_FILTER("last_fres: %.2f Hz  |  target_fres: %.2f Hz  |---|  last_q: %.4f  |  target_q: %.4f", iir_filter->last_fres, iir_filter->target_fres, iir_filter->last_q, iir_filter->target_q);
-
-                fluid_iir_filter_calculate_coefficients(iir_filter, output_rate, &dsp_a1, &dsp_a2, &dsp_b02, &dsp_b1);
-            }
         }
 
         iir_filter->hist1 = dsp_hist1;
@@ -131,13 +103,11 @@ fluid_iir_filter_apply(fluid_iir_filter_t *iir_filter,
         iir_filter->b02 = dsp_b02;
         iir_filter->b1 = dsp_b1;
 
-        iir_filter->fres_incr_count = fres_incr_count;
-        iir_filter->q_incr_count = q_incr_count;
-
         fluid_check_fpe("voice_filter");
     }
 }
 
+
 DECLARE_FLUID_RVOICE_FUNCTION(fluid_iir_filter_init)
 {
     fluid_iir_filter_t *iir_filter = obj;
@@ -159,7 +129,7 @@ fluid_iir_filter_reset(fluid_iir_filter_t *iir_filter)
     iir_filter->hist1 = 0;
     iir_filter->hist2 = 0;
     iir_filter->last_fres = -1.;
-    iir_filter->last_q = 0;
+    iir_filter->q_lin = 0;
     iir_filter->filter_startup = 1;
 }
 
@@ -169,8 +139,7 @@ DECLARE_FLUID_RVOICE_FUNCTION(fluid_iir_filter_set_fres)
     fluid_real_t fres = param[0].real;
 
     iir_filter->fres = fres;
-
-    LOG_FILTER("fluid_iir_filter_set_fres: fres= %f [acents]",fres);
+    iir_filter->last_fres = -1.;
 }
 
 static fluid_real_t fluid_iir_filter_q_from_dB(fluid_real_t q_dB)
@@ -182,7 +151,6 @@ static fluid_real_t fluid_iir_filter_q_from_dB(fluid_real_t q_dB)
     /* Range: SF2.01 section 8.1.3 # 8 (convert from cB to dB => /10) */
     fluid_clip(q_dB, 0.0f, 96.0f);
 
-#if 0 /* unused in Wine */
     /* Short version: Modify the Q definition in a way, that a Q of 0
      * dB leads to no resonance hump in the freq. response.
      *
@@ -199,7 +167,6 @@ static fluid_real_t fluid_iir_filter_q_from_dB(fluid_real_t q_dB)
      * response of a non-resonant filter.  This idea is implemented as
      * follows: */
     q_dB -= 3.01f;
-#endif /* unused in Wine */
 
     /* The 'sound font' Q is defined in dB. The filter needs a linear
        q. Convert. */
@@ -212,55 +179,58 @@ DECLARE_FLUID_RVOICE_FUNCTION(fluid_iir_filter_set_q)
     fluid_real_t q = param[0].real;
     int flags = iir_filter->flags;
 
-    LOG_FILTER("fluid_iir_filter_set_q: Q= %f [dB]",q);
-
     if(flags & FLUID_IIR_Q_ZERO_OFF && q <= 0.0)
     {
         q = 0;
     }
     else if(flags & FLUID_IIR_Q_LINEAR)
     {
-        /* q is linear (only for user-defined filter) */
+        /* q is linear (only for user-defined filter)
+         * increase to avoid Q being somewhere between zero and one,
+         * which results in some strange amplified lowpass signal
+         */
+        q++;
     }
     else
     {
         q = fluid_iir_filter_q_from_dB(q);
     }
 
-    LOG_FILTER("fluid_iir_filter_set_q: Q= %f [linear]",q);
+    iir_filter->q_lin = q;
+    iir_filter->filter_gain = 1.0;
 
-    if(iir_filter->filter_startup)
+    if(!(flags & FLUID_IIR_NO_GAIN_AMP))
     {
-        iir_filter->last_q = q;
-    }
-    else
-    {
-        static const fluid_real_t q_incr_count = FLUID_BUFSIZE;
-        iir_filter->q_incr = (q - iir_filter->last_q) / (q_incr_count);
-        iir_filter->q_incr_count = q_incr_count;
+        /* SF 2.01 page 59:
+         *
+         *  The SoundFont specs ask for a gain reduction equal to half the
+         *  height of the resonance peak (Q).  For example, for a 10 dB
+         *  resonance peak, the gain is reduced by 5 dB.  This is done by
+         *  multiplying the total gain with sqrt(1/Q).  `Sqrt' divides dB
+         *  by 2 (100 lin = 40 dB, 10 lin = 20 dB, 3.16 lin = 10 dB etc)
+         *  The gain is later factored into the 'b' coefficients
+         *  (numerator of the filter equation).  This gain factor depends
+         *  only on Q, so this is the right place to calculate it.
+         */
+        iir_filter->filter_gain /= FLUID_SQRT(q);
     }
-#ifdef DBG_FILTER
-    iir_filter->target_q = q;
-#endif
+
+    /* The synthesis loop will have to recalculate the filter coefficients. */
+    iir_filter->last_fres = -1.;
 }
 
 static FLUID_INLINE void
 fluid_iir_filter_calculate_coefficients(fluid_iir_filter_t *iir_filter,
-                                        fluid_real_t output_rate,
-                                        fluid_real_t *a1_out, fluid_real_t *a2_out,
-                                        fluid_real_t *b02_out, fluid_real_t *b1_out)
+                                        int transition_samples,
+                                        fluid_real_t output_rate)
 {
-    // FLUID_IIR_Q_LINEAR may switch the filter off by setting Q==0
-    // Due to the linear smoothing, last_q may not exactly become zero.
-    if(FLUID_FABS(iir_filter->last_q) <= 0.001)
+    /* FLUID_IIR_Q_LINEAR may switch the filter off by setting Q==0 */
+    if(iir_filter->q_lin == 0)
     {
         return;
     }
     else
     {
-        int flags = iir_filter->flags;
-        fluid_real_t filter_gain = 1.0f;
-
         /*
          * Those equations from Robert Bristow-Johnson's `Cookbook
          * formulae for audio EQ biquad filter coefficients', obtained
@@ -274,7 +244,7 @@ fluid_iir_filter_calculate_coefficients(fluid_iir_filter_t *iir_filter,
                                             (iir_filter->last_fres / output_rate);
         fluid_real_t sin_coeff = FLUID_SIN(omega);
         fluid_real_t cos_coeff = FLUID_COS(omega);
-        fluid_real_t alpha_coeff = sin_coeff / (2.0f * iir_filter->last_q);
+        fluid_real_t alpha_coeff = sin_coeff / (2.0f * iir_filter->q_lin);
         fluid_real_t a0_inv = 1.0f / (1.0f + alpha_coeff);
 
         /* Calculate the filter coefficients. All coefficients are
@@ -282,35 +252,20 @@ fluid_iir_filter_calculate_coefficients(fluid_iir_filter_t *iir_filter,
          *
          * Here a couple of multiplications are saved by reusing common expressions.
          * The original equations should be:
-         *  iir_filter->b0=(1.-cos_coeff)*a0_inv*0.5*filter_gain;
-         *  iir_filter->b1=(1.-cos_coeff)*a0_inv*filter_gain;
-         *  iir_filter->b2=(1.-cos_coeff)*a0_inv*0.5*filter_gain; */
+         *  iir_filter->b0=(1.-cos_coeff)*a0_inv*0.5*iir_filter->filter_gain;
+         *  iir_filter->b1=(1.-cos_coeff)*a0_inv*iir_filter->filter_gain;
+         *  iir_filter->b2=(1.-cos_coeff)*a0_inv*0.5*iir_filter->filter_gain; */
 
         /* "a" coeffs are same for all 3 available filter types */
         fluid_real_t a1_temp = -2.0f * cos_coeff * a0_inv;
         fluid_real_t a2_temp = (1.0f - alpha_coeff) * a0_inv;
-        fluid_real_t b02_temp, b1_temp;
 
-        if(!(flags & FLUID_IIR_NO_GAIN_AMP))
-        {
-            /* SF 2.01 page 59:
-            *
-            *  The SoundFont specs ask for a gain reduction equal to half the
-            *  height of the resonance peak (Q).  For example, for a 10 dB
-            *  resonance peak, the gain is reduced by 5 dB.  This is done by
-            *  multiplying the total gain with sqrt(1/Q).  `Sqrt' divides dB
-            *  by 2 (100 lin = 40 dB, 10 lin = 20 dB, 3.16 lin = 10 dB etc)
-            *  The gain is later factored into the 'b' coefficients
-            *  (numerator of the filter equation).  This gain factor depends
-            *  only on Q, so this is the right place to calculate it.
-            */
-            filter_gain /= FLUID_SQRT(iir_filter->last_q);
-        }
+        fluid_real_t b02_temp, b1_temp;
 
         switch(iir_filter->type)
         {
         case FLUID_IIR_HIGHPASS:
-            b1_temp = (1.0f + cos_coeff) * a0_inv * filter_gain;
+            b1_temp = (1.0f + cos_coeff) * a0_inv * iir_filter->filter_gain;
 
             /* both b0 -and- b2 */
             b02_temp = b1_temp * 0.5f;
@@ -319,7 +274,7 @@ fluid_iir_filter_calculate_coefficients(fluid_iir_filter_t *iir_filter,
             break;
 
         case FLUID_IIR_LOWPASS:
-            b1_temp = (1.0f - cos_coeff) * a0_inv * filter_gain;
+            b1_temp = (1.0f - cos_coeff) * a0_inv * iir_filter->filter_gain;
 
             /* both b0 -and- b2 */
             b02_temp = b1_temp * 0.5f;
@@ -330,10 +285,11 @@ fluid_iir_filter_calculate_coefficients(fluid_iir_filter_t *iir_filter,
             return;
         }
 
-        *a1_out = a1_temp;
-        *a2_out = a2_temp;
-        *b02_out = b02_temp;
-        *b1_out = b1_temp;
+        iir_filter->a1 = a1_temp;
+        iir_filter->a2 = a2_temp;
+        iir_filter->b02 = b02_temp;
+        iir_filter->b1 = b1_temp;
+        iir_filter->filter_startup = 0;
 
         fluid_check_fpe("voice_write filter calculation");
     }
@@ -344,13 +300,7 @@ void fluid_iir_filter_calc(fluid_iir_filter_t *iir_filter,
                            fluid_real_t output_rate,
                            fluid_real_t fres_mod)
 {
-    unsigned int calc_coeff_flag = FALSE;
-    fluid_real_t fres, fres_diff;
-
-    if(iir_filter->type == FLUID_IIR_DISABLED)
-    {
-        return;
-    }
+    fluid_real_t fres;
 
     /* calculate the frequency of the resonant filter in Hz */
     fres = fluid_ct2hz(iir_filter->fres + fres_mod);
@@ -374,47 +324,21 @@ void fluid_iir_filter_calc(fluid_iir_filter_t *iir_filter,
         fres = 5.f;
     }
 
-    LOG_FILTER("%f + %f = %f cents = %f Hz | Q: %f", iir_filter->fres, fres_mod, iir_filter->fres + fres_mod, fres, iir_filter->last_q);
-
+    // FLUID_LOG(FLUID_INFO, "%f + %f cents = %f cents = %f Hz | Q: %f", iir_filter->fres, fres_mod, iir_filter->fres + fres_mod, fres, iir_filter->q_lin);
     /* if filter enabled and there is a significant frequency change.. */
-    fres_diff = fres - iir_filter->last_fres;
-    if(iir_filter->filter_startup)
+    if(iir_filter->type != FLUID_IIR_DISABLED && FLUID_FABS(fres - iir_filter->last_fres) > 0.01f)
     {
-        // The filer was just starting up, make sure to calculate initial coefficients for the initial Q value, even though the fres may not have changed
-        calc_coeff_flag = TRUE;
-
-        iir_filter->fres_incr_count = 0;
+        /* The filter coefficients have to be recalculated (filter
+         * parameters have changed). Recalculation for various reasons is
+         * forced by setting last_fres to -1.  The flag filter_startup
+         * indicates, that the DSP loop runs for the first time, in this
+         * case, the filter is set directly, instead of smoothly fading
+         * between old and new settings. */
         iir_filter->last_fres = fres;
-        iir_filter->filter_startup = 0;
-    }
-    else if(FLUID_FABS(fres_diff) > 0.01f)
-    {
-        fluid_real_t fres_incr_count = FLUID_BUFSIZE;
-        fluid_real_t num_buffers = iir_filter->last_q;
-        fluid_clip(num_buffers, 1, 5);
-        // For high values of Q, the phase gets really steep. To prevent clicks when quickly modulating fres in this case, we need to smooth out "slower".
-        // This is done by simply using Q times FLUID_BUFSIZE samples for the interpolation to complete, capped at 5.
-        // 5 was chosen because the phase doesn't really get any steeper when continuing to increase Q.
-        fres_incr_count *= num_buffers;
-        iir_filter->fres_incr = fres_diff / (fres_incr_count);
-        iir_filter->fres_incr_count = fres_incr_count;
-#ifdef DBG_FILTER
-        iir_filter->target_fres = fres;
-#endif
-
-        // The filter coefficients have to be recalculated (filter cutoff has changed).
-        calc_coeff_flag = TRUE;
-    }
-    else
-    {
-        // We do not account for any change of Q here - if it was changed q_incro_count will be non-zero and recalculating the coeffs
-        // will be taken care of in fluid_iir_filter_apply().
+        fluid_iir_filter_calculate_coefficients(iir_filter, FLUID_BUFSIZE,
+                                                output_rate);
     }
 
-    if(calc_coeff_flag)
-    {
-        fluid_iir_filter_calculate_coefficients(iir_filter, output_rate, &iir_filter->a1, &iir_filter->a2, &iir_filter->b02, &iir_filter->b1);
-    }
 
     fluid_check_fpe("voice_write DSP coefficients");
 
