diff --git a/libs/fluidsynth/src/rvoice/fluid_rvoice.c b/libs/fluidsynth/src/rvoice/fluid_rvoice.c
index e07896fea01ae6f88ac3ec43acf8b01b81ed95e3..3db52b60fb77ce6def8625cdc0879f95354988a4 100644
--- a/libs/fluidsynth/src/rvoice/fluid_rvoice.c
+++ b/libs/fluidsynth/src/rvoice/fluid_rvoice.c
@@ -435,22 +435,6 @@ fluid_rvoice_write(fluid_rvoice_t *voice, fluid_real_t *dsp_buf)
                  || (voice->dsp.samplemode == FLUID_LOOP_UNTIL_RELEASE
                      && fluid_adsr_env_get_section(&voice->envlfo.volenv) < FLUID_VOICE_ENVRELEASE);
 
-    /*************** resonant filter ******************/
-    // Only "prepare" the filter here, the filter itself will be applied in the dsp_interpolation routines below.
-    // This is to satisfy SF2 Section 9.1.8, particularly, the filtered output must be gain-adjusted by the volEnv.
-    // Applying the filter after applying the gain from the volEnv might cause audible clicks for when turning off
-    // voices that are filtered by a high Q, see https://github.com/FluidSynth/fluidsynth/issues/1427
-    //
-    // Note that at this point we are using voice->dsp.output_rate which is set to the synth's output rate, because
-    // the filter will receive the interpolated waveform.
-
-    fluid_iir_filter_calc(&voice->resonant_filter, voice->dsp.output_rate,
-                          fluid_lfo_get_val(&voice->envlfo.modlfo) * voice->envlfo.modlfo_to_fc +
-                          modenv_val * voice->envlfo.modenv_to_fc);
-
-    /* additional custom filter - only uses the fixed modulator, no lfos... */
-    fluid_iir_filter_calc(&voice->resonant_custom_filter, voice->dsp.output_rate, 0);
-
     /*********************** run the dsp chain ************************
      * The sample is mixed with the output buffer.
      * The buffer has to be filled from 0 to FLUID_BUFSIZE-1.
@@ -465,26 +449,28 @@ fluid_rvoice_write(fluid_rvoice_t *voice, fluid_real_t *dsp_buf)
         //
         // Currently, this does access the sample buffers, which is redundant and could be optimized away.
         // On the other hand, entering this if-clause is not supposed to happen often.
-        return fluid_rvoice_dsp_interpolate_none(voice, dsp_buf, is_looping);
+        //
+        // Also note, that we're returning directly without running the IIR filter below.
+        return fluid_rvoice_dsp_interpolate_none(&voice->dsp, dsp_buf, is_looping);
     }
 
     switch(voice->dsp.interp_method)
     {
     case FLUID_INTERP_NONE:
-        count = fluid_rvoice_dsp_interpolate_none(voice, dsp_buf, is_looping);
+        count = fluid_rvoice_dsp_interpolate_none(&voice->dsp, dsp_buf, is_looping);
         break;
 
     case FLUID_INTERP_LINEAR:
-        count = fluid_rvoice_dsp_interpolate_linear(voice, dsp_buf, is_looping);
+        count = fluid_rvoice_dsp_interpolate_linear(&voice->dsp, dsp_buf, is_looping);
         break;
 
     case FLUID_INTERP_4THORDER:
     default:
-        count = fluid_rvoice_dsp_interpolate_4th_order(voice, dsp_buf, is_looping);
+        count = fluid_rvoice_dsp_interpolate_4th_order(&voice->dsp, dsp_buf, is_looping);
         break;
 
     case FLUID_INTERP_7THORDER:
-        count = fluid_rvoice_dsp_interpolate_7th_order(voice, dsp_buf, is_looping);
+        count = fluid_rvoice_dsp_interpolate_7th_order(&voice->dsp, dsp_buf, is_looping);
         break;
     }
 
@@ -496,6 +482,18 @@ fluid_rvoice_write(fluid_rvoice_t *voice, fluid_real_t *dsp_buf)
         return count;
     }
 
+    /*************** resonant filter ******************/
+
+    fluid_iir_filter_calc(&voice->resonant_filter, voice->dsp.output_rate,
+                          fluid_lfo_get_val(&voice->envlfo.modlfo) * voice->envlfo.modlfo_to_fc +
+                          modenv_val * voice->envlfo.modenv_to_fc);
+
+    fluid_iir_filter_apply(&voice->resonant_filter, dsp_buf, count);
+
+    /* additional custom filter - only uses the fixed modulator, no lfos... */
+    fluid_iir_filter_calc(&voice->resonant_custom_filter, voice->dsp.output_rate, 0);
+    fluid_iir_filter_apply(&voice->resonant_custom_filter, dsp_buf, count);
+
     return count;
 }
 
