diff --git a/libs/fluidsynth/src/synth/fluid_voice.c b/libs/fluidsynth/src/synth/fluid_voice.c
index c1daf3d316d554d36c35ad92eae34bed2c088e37..ca0bad05e9c226ec0d4c711f26421f275c39f691 100644
--- a/libs/fluidsynth/src/synth/fluid_voice.c
+++ b/libs/fluidsynth/src/synth/fluid_voice.c
@@ -859,6 +859,11 @@ fluid_voice_update_param(fluid_voice_t *voice, int gen)
         break;
 
     case GEN_FILTERFC:
+        /* The resonance frequency is converted from absolute cents to
+         * midicents .val and .mod are both used, this permits real-time
+         * modulation.  The allowed range is tested in the 'fluid_ct2hz'
+         * function [PH,20021214]
+         */
         UPDATE_RVOICE_GENERIC_R1(fluid_iir_filter_set_fres, &voice->rvoice->resonant_filter, x);
         break;
 
@@ -1398,7 +1403,6 @@ fluid_voice_kill_excl(fluid_voice_t *voice)
     */
     fluid_voice_gen_set(voice, GEN_EXCLUSIVECLASS, 0);
 
-#if 0 /* unused in Wine */
     /* Speed up the volume envelope */
     /* The value was found through listening tests with hi-hat samples. */
     fluid_voice_gen_set(voice, GEN_VOLENVRELEASE, -200);
@@ -1407,12 +1411,6 @@ fluid_voice_kill_excl(fluid_voice_t *voice)
     /* Speed up the modulation envelope */
     fluid_voice_gen_set(voice, GEN_MODENVRELEASE, -200);
     fluid_voice_update_param(voice, GEN_MODENVRELEASE);
-        gen[i].val = fluid_gen_info[i].def;
-#else
-    /* Speed up the volume envelope */
-    fluid_voice_gen_set(voice, GEN_VOLENVRELEASE, -32768);
-    fluid_voice_update_param(voice, GEN_VOLENVRELEASE);
-#endif
 
     at_tick = fluid_channel_get_min_note_length_ticks(voice->channel);
     UPDATE_RVOICE_I1(fluid_rvoice_noteoff, at_tick);
