diff --git a/libs/fluidsynth/src/utils/fluid_sys.h b/libs/fluidsynth/src/utils/fluid_sys.h
index 79aeef0fb90996f4cea2d38c85e08af24730529e..15f3f57b7566a06054960861c5d92e93dc33f47a 100644
--- a/libs/fluidsynth/src/utils/fluid_sys.h
+++ b/libs/fluidsynth/src/utils/fluid_sys.h
@@ -156,13 +156,6 @@ typedef gintptr  intptr_t;
 #pragma warning(disable : 4996)
 #endif
 
-/*
- * Required by Windows-specific sf_wchar_open() from old libsndfile
- * versions before v1.1.0, that takes a UTF16_BE encoded filename.
- * Note that FluidSynth needs libsndfile >= v1.2.1 anyway.
- */
-#define ENABLE_SNDFILE_WINDOWS_PROTOTYPES 1
-
 #endif
 
 /* Darwin special defines (taken from config_macosx.h) */
