diff --git a/libs/lcms2/src/cmserr.c b/libs/lcms2/src/cmserr.c
index 2603b876438b80b360052227a6de5b3f5f28114e..a0b338491a416b08837abcc5a63bc1c1caa6c9fa 100644
--- a/libs/lcms2/src/cmserr.c
+++ b/libs/lcms2/src/cmserr.c
@@ -1,7 +1,7 @@
 //---------------------------------------------------------------------------------
 //
 //  Little Color Management System
-//  Copyright (c) 1998-2024 Marti Maria Saguer
+//  Copyright (c) 1998-2023 Marti Maria Saguer
 //
 // Permission is hereby granted, free of charge, to any person obtaining
 // a copy of this software and associated documentation files (the "Software"),
