diff --git a/libs/lcms2/src/cmsgamma.c b/libs/lcms2/src/cmsgamma.c
index 3a8352d72d283c9028f71a351397cb7c8c1ace4b..cb0abd45db203f489b8c9cf0a2f6a76409fd3ac9 100644
--- a/libs/lcms2/src/cmsgamma.c
+++ b/libs/lcms2/src/cmsgamma.c
@@ -1,7 +1,7 @@
 //---------------------------------------------------------------------------------
 //
 //  Little Color Management System
-//  Copyright (c) 1998-2024 Marti Maria Saguer
+//  Copyright (c) 1998-2023 Marti Maria Saguer
 //
 // Permission is hereby granted, free of charge, to any person obtaining
 // a copy of this software and associated documentation files (the "Software"),
