diff --git a/libs/lcms2/src/cmsgmt.c b/libs/lcms2/src/cmsgmt.c
index b4003e55bb97b048a1ff638d0eda980d97ded0ae..52a0829a42d495b36200a7ea96aee76b315e317d 100644
--- a/libs/lcms2/src/cmsgmt.c
+++ b/libs/lcms2/src/cmsgmt.c
@@ -1,7 +1,7 @@
 //---------------------------------------------------------------------------------
 //
 //  Little Color Management System
-//  Copyright (c) 1998-2024 Marti Maria Saguer
+//  Copyright (c) 1998-2021 Marti Maria Saguer
 //
 // Permission is hereby granted, free of charge, to any person obtaining
 // a copy of this software and associated documentation files (the "Software"),
@@ -298,9 +298,8 @@ cmsPipeline* _cmsCreateGamutCheckPipeline(cmsContext ContextID,
     cmsUInt32Number dwFormat;
     GAMUTCHAIN Chain;
     cmsUInt32Number nGridpoints;
-    cmsInt32Number nChannels, nInputChannels;
+    cmsInt32Number nChannels;
     cmsColorSpaceSignature ColorSpace;
-    cmsColorSpaceSignature InputColorSpace;
     cmsUInt32Number i;
     cmsHPROFILE ProfileList[256];
     cmsBool     BPCList[256];
@@ -346,13 +345,11 @@ cmsPipeline* _cmsCreateGamutCheckPipeline(cmsContext ContextID,
     AdaptationList[nGamutPCSposition] = 1.0;
     IntentList[nGamutPCSposition] = INTENT_RELATIVE_COLORIMETRIC;
 
+
     ColorSpace  = cmsGetColorSpace(hGamut);
     nChannels   = cmsChannelsOfColorSpace(ColorSpace);
     nGridpoints = _cmsReasonableGridpointsByColorspace(ColorSpace, cmsFLAGS_HIGHRESPRECALC);
-
-    InputColorSpace = cmsGetColorSpace(ProfileList[0]);
-    nInputChannels  = cmsChannelsOfColorSpace(InputColorSpace);
-    dwFormat        = (CHANNELS_SH(nInputChannels)|BYTES_SH(2));
+    dwFormat    = (CHANNELS_SH(nChannels)|BYTES_SH(2));
 
     // 16 bits to Lab double
     Chain.hInput = cmsCreateExtendedTransform(ContextID,
