diff --git a/libs/lcms2/src/cmshalf.c b/libs/lcms2/src/cmshalf.c
index f15230873ba9bd368cc024da6ea08a52b2121668..043ee530b46a60f21736a93149359ae4a998d841 100644
--- a/libs/lcms2/src/cmshalf.c
+++ b/libs/lcms2/src/cmshalf.c
@@ -1,7 +1,7 @@
 //---------------------------------------------------------------------------------
 //
 //  Little Color Management System
-//  Copyright (c) 1998-2024 Marti Maria Saguer
+//  Copyright (c) 1998-2023 Marti Maria Saguer
 //
 // Permission is hereby granted, free of charge, to any person obtaining
 // a copy of this software and associated documentation files (the "Software"),
