diff --git a/libs/lcms2/src/cmsio0.c b/libs/lcms2/src/cmsio0.c
index a4a7397be5bd9939c3c07ddbe6add2051883cb00..92aa00a581f3aab965e1ea743447721de690eb8c 100644
--- a/libs/lcms2/src/cmsio0.c
+++ b/libs/lcms2/src/cmsio0.c
@@ -1,7 +1,7 @@
 //---------------------------------------------------------------------------------
 //
 //  Little Color Management System
-//  Copyright (c) 1998-2024 Marti Maria Saguer
+//  Copyright (c) 1998-2023 Marti Maria Saguer
 //
 // Permission is hereby granted, free of charge, to any person obtaining
 // a copy of this software and associated documentation files (the "Software"),
@@ -277,11 +277,6 @@ cmsIOHANDLER* CMSEXPORT cmsOpenIOhandlerFromMem(cmsContext ContextID, void *Buff
         fm = (FILEMEM*) _cmsMallocZero(ContextID, sizeof(FILEMEM));
         if (fm == NULL) goto Error;
 
-        if (Buffer == NULL) {
-            cmsSignalError(ContextID, cmsERROR_WRITE, "Couldn't write profile to NULL pointer");
-            goto Error;
-        }
-
         fm ->Block = (cmsUInt8Number*) Buffer;
         fm ->FreeBlockOnClose = FALSE;
         fm ->Size    = size;
