diff --git a/libs/lcms2/src/cmslut.c b/libs/lcms2/src/cmslut.c
index 6842884ff4f7a494db74609592c3ea2f3e9ab7d8..1ad5ab3b9af4944531a657263a1f890b2a6ab90e 100644
--- a/libs/lcms2/src/cmslut.c
+++ b/libs/lcms2/src/cmslut.c
@@ -1,7 +1,7 @@
 //---------------------------------------------------------------------------------
 //
 //  Little Color Management System
-//  Copyright (c) 1998-2024 Marti Maria Saguer
+//  Copyright (c) 1998-2023 Marti Maria Saguer
 //
 // Permission is hereby granted, free of charge, to any person obtaining
 // a copy of this software and associated documentation files (the "Software"),
@@ -1080,7 +1080,7 @@ cmsStage* _cmsStageNormalizeFromLabFloat(cmsContext ContextID)
     return mpe;
 }
 
-// From XYZ to floating point PCS
+// Fom XYZ to floating point PCS
 cmsStage* _cmsStageNormalizeFromXyzFloat(cmsContext ContextID)
 {
 #define n (32768.0/65535.0)
