diff --git a/libs/lcms2/src/cmssamp.c b/libs/lcms2/src/cmssamp.c
index ab1a3abbf1bf54fb315fdd7632c25a94438568d4..e0b3bd003f819a9f2446694b7fab104ad038e756 100644
--- a/libs/lcms2/src/cmssamp.c
+++ b/libs/lcms2/src/cmssamp.c
@@ -1,7 +1,7 @@
 //---------------------------------------------------------------------------------
 //
 //  Little Color Management System
-//  Copyright (c) 1998-2024 Marti Maria Saguer
+//  Copyright (c) 1998-2023 Marti Maria Saguer
 //
 // Permission is hereby granted, free of charge, to any person obtaining
 // a copy of this software and associated documentation files (the "Software"),
