diff --git a/libs/lcms2/src/cmswtpnt.c b/libs/lcms2/src/cmswtpnt.c
index ed2e5491f281af6d23d27a4d024e533ef438a49f..a73eaa7e366e7194e30d9a8c1380f113b7658c97 100644
--- a/libs/lcms2/src/cmswtpnt.c
+++ b/libs/lcms2/src/cmswtpnt.c
@@ -1,7 +1,7 @@
 //---------------------------------------------------------------------------------
 //
 //  Little Color Management System
-//  Copyright (c) 1998-2024 Marti Maria Saguer
+//  Copyright (c) 1998-2023 Marti Maria Saguer
 //
 // Permission is hereby granted, free of charge, to any person obtaining
 // a copy of this software and associated documentation files (the "Software"),
@@ -349,3 +349,5 @@ cmsBool CMSEXPORT cmsAdaptToIlluminant(cmsCIEXYZ* Result,
 
     return TRUE;
 }
+
+
