diff --git a/libs/mpg123/config.h b/libs/mpg123/config.h
index b7a117c83fc7f19f9cd6a4ed1d2b696efdf9324f..d9f163fd2a3f2ddc241a88f7ce9a773edd562571 100644
--- a/libs/mpg123/config.h
+++ b/libs/mpg123/config.h
@@ -245,12 +245,6 @@
 /* Define to 1 if you have the <string.h> header file. */
 #define HAVE_STRING_H 1
 
-/* Define to 1 if you have the `strtok_r' function. */
-/* #undef HAVE_STRTOK_R */
-
-/* Define to 1 if you have the `strtok_s' function. */
-#define HAVE_STRTOK_S 1
-
 /* Define to 1 if you have the <sun/audioio.h> header file. */
 /* #undef HAVE_SUN_AUDIOIO_H */
 
@@ -445,7 +439,7 @@
 #define PACKAGE_NAME "mpg123"
 
 /* Define to the full name and version of this package. */
-#define PACKAGE_STRING "mpg123 1.33.0"
+#define PACKAGE_STRING "mpg123 1.32.9"
 
 /* Define to the one symbol short name of this package. */
 #define PACKAGE_TARNAME "mpg123"
@@ -454,7 +448,7 @@
 #define PACKAGE_URL ""
 
 /* Define to the version of this package. */
-#define PACKAGE_VERSION "1.33.0"
+#define PACKAGE_VERSION "1.32.9"
 
 /* Define to only include portable library API (no off_t, no internal I/O). */
 /* #undef PORTABLE_API */
@@ -486,7 +480,7 @@
 /* #undef USE_YASM_FOR_AVX */
 
 /* Version number of package */
-#define VERSION "1.33.0"
+#define VERSION "1.32.9"
 
 /* Define to use Win32 named pipes */
 #define WANT_WIN32_FIFO 1
