diff --git a/libs/mpg123/src/compat/compat.h b/libs/mpg123/src/compat/compat.h
index 0a16fd3281075c9cc106663ef586c2f6ad272fd9..23df921b113680cecd294c49c74bc3dc3e6c3782 100644
--- a/libs/mpg123/src/compat/compat.h
+++ b/libs/mpg123/src/compat/compat.h
@@ -110,17 +110,8 @@
 
 typedef unsigned char byte;
 
-// Annoying hackery to select a safe strtok variant. MS decided to call their strtok_r strtok_s, while
-// C11 declares another strtok_s with different prototype. Thanks to you all.
-#ifdef HAVE_STRTOK_R
-#define INT123_compat_strtok(a, b, c) strtok_r((a), (b), (c))
-#endif
-
 #if (defined(_UCRT) || defined(_MSC_VER) || (defined(__MINGW32__) || defined(__MINGW64__)) || (defined(__WATCOMC__) && defined(__NT__))) && !defined(__CYGWIN__)
 #define MPG123_COMPAT_MSVCRT_IO
-#ifndef INT123_compat_strtok
-#define INT123_compat_strtok(a, b, c) strtok_s((a), (b), (c))
-#endif
 #endif
 
 #if defined(MPG123_COMPAT_MSVCRT_IO)
@@ -159,11 +150,6 @@ typedef unsigned char byte;
 #include <io.h>
 #endif
 
-#ifndef INT123_compat_strtok
-#warning "no safe strtok found"
-#define INT123_compat_strtok(a, b, c) strtok((a), (b))
-#endif
-
 /* A safe realloc also for very old systems where realloc(NULL, size) returns NULL. */
 void *INT123_safe_realloc(void *ptr, size_t size);
 // Also freeing ptr if result is NULL. You can do
