diff --git a/libs/mpg123/src/libmpg123/id3.c b/libs/mpg123/src/libmpg123/id3.c
index 0bd5cda70b6e5c42e6bf64fec08bf81589866507..180f489fdfdd5a6fcffa27eaaf756c960f90bf68 100644
--- a/libs/mpg123/src/libmpg123/id3.c
+++ b/libs/mpg123/src/libmpg123/id3.c
@@ -794,7 +794,7 @@ static int promote_framename(mpg123_handle *fr, char *id) /* fr because of VERBO
 
 #endif /* NO_ID3V2 */
 
-static int store_id3v2( mpg123_handle *fr
+int store_id3v2( mpg123_handle *fr
 ,	unsigned long first4bytes, unsigned char buf[6], unsigned long length )
 {
 	int ret = 1;
