diff --git a/libs/mpg123/src/libmpg123/libmpg123.c b/libs/mpg123/src/libmpg123/libmpg123.c
index 22b3d09dcdadf4e0569fd8fc79d19863904f91e9..971e503e3347d317e26739cb068c6c2a4f6a58b7 100644
--- a/libs/mpg123/src/libmpg123/libmpg123.c
+++ b/libs/mpg123/src/libmpg123/libmpg123.c
@@ -15,7 +15,9 @@
 #define FORCE_ACCURATE
 #include "../common/sample.h"
 #include "parse.h"
+#ifndef PORTABLE_API
 #include "lfs_wrap.h"
+#endif
 
 #include "../common/debug.h"
 
@@ -557,23 +559,6 @@ double attribute_align_arg mpg123_geteq2(mpg123_handle *mh, int channel, int ban
 	return mpg123_geteq(mh, channel, band);
 }
 
-// LFS wrapper code is so agnostic to it all now that internal I/O is portable
-// as long as you do not mix in off_t API.
-int attribute_align_arg mpg123_open64(mpg123_handle *mh, const char *path)
-{
-	if(mh == NULL) return MPG123_BAD_HANDLE;
-
-	mpg123_close(mh);
-	if(!path)
-		return MPG123_ERR;
-	// sets callbacks, only allocating wrapperdata handle if internal callbacks involved
-	int ret = INT123_wrap_open( mh, NULL, path, -1
-	,	mh->p.timeout, mh->p.flags & MPG123_QUIET );
-	if(!ret)
-		ret = INT123_open_stream_handle(mh, mh->wrapperdata);
-	return ret;
-}
-
 #ifndef PORTABLE_API
 
 #ifdef FORCED_OFF_64
@@ -584,13 +569,21 @@ int attribute_align_arg mpg123_open64(mpg123_handle *mh, const char *path)
 #define mpg123_open_handle mpg123_open_handle_64
 #endif
 
-// This now is agnostic to off_t choice, but still subject to renaming
-// for legacy reasons.
+/* plain file access, no http! */
 int attribute_align_arg mpg123_open(mpg123_handle *mh, const char *path)
 {
-	return mpg123_open64(mh, path);
+	if(mh == NULL) return MPG123_BAD_HANDLE;
+
+	mpg123_close(mh);
+	if(!path)
+		return MPG123_ERR;
+	// sets callbacks, only allocating wrapperdata handle if internal callbacks involved
+	int ret = INT123_wrap_open( mh, NULL, path, -1
+	,	mh->p.timeout, mh->p.flags & MPG123_QUIET );
+	if(!ret)
+		ret = INT123_open_stream_handle(mh, mh->wrapperdata);
+	return ret;
 }
-#endif // PORTABLE_API
 
 // The convenience function mpg123_open_fixed() wraps over acual mpg123_open
 // and hence needs to have the exact same code in lfs_wrap.c. The flesh is
@@ -631,28 +624,17 @@ static int INT123_open_fixed_post(mpg123_handle *mh, int channels, int encoding)
 	return err;
 }
 
-int attribute_align_arg mpg123_open_fixed64( mpg123_handle *mh, const char *path
+int attribute_align_arg mpg123_open_fixed( mpg123_handle *mh, const char *path
 ,	int channels, int encoding )
 {
 	int err = INT123_open_fixed_pre(mh, channels, encoding);
 	if(err == MPG123_OK)
-		err = mpg123_open64(mh, path);
+		err = mpg123_open(mh, path);
 	if(err == MPG123_OK)
 		err = INT123_open_fixed_post(mh, channels, encoding);
 	return err;
 }
 
-#ifndef PORTABLE_API
-// Only to have the modern offset-agnostic open under a fixed name.
-int attribute_align_arg mpg123_open_fixed( mpg123_handle *mh, const char *path
-,	int channels, int encoding )
-{
-	return mpg123_open_fixed64(mh, path, channels, encoding);
-}
-
-// Won't define a 'portable' variant of this, as I cannot guess
-// properties of the handed-in fd, which in theory, on specific platforms,
-// could not support large files.
 int attribute_align_arg mpg123_open_fd(mpg123_handle *mh, int fd)
 {
 	if(mh == NULL) return MPG123_BAD_HANDLE;
@@ -668,33 +650,21 @@ int attribute_align_arg mpg123_open_fd(mpg123_handle *mh, int fd)
 }
 #endif // PORTABLE_API
 
-// Only works with int64 reader setup.
-int attribute_align_arg mpg123_open_handle64(mpg123_handle *mh, void *iohandle)
-{
-	if(mh == NULL) return MPG123_BAD_HANDLE;
-
-	mpg123_close(mh);
-	return INT123_open_stream_handle(mh, iohandle);
-}
-
-#ifndef PORTABLE_API
-// Change from 1.32: No largefile-renamed symbols in a library with strict
-// portable API.
-// I allow that breaking change since this is far from a standard libmpg123 build.
 int attribute_align_arg mpg123_open_handle(mpg123_handle *mh, void *iohandle)
 {
 	if(mh == NULL) return MPG123_BAD_HANDLE;
 
 	mpg123_close(mh);
 	int ret;
+#ifndef PORTABLE_API
 	ret = INT123_wrap_open( mh, iohandle, NULL, -1
 	,	mh->p.timeout, mh->p.flags & MPG123_QUIET );
 	iohandle = ret == LFS_WRAP_NONE ? iohandle : mh->wrapperdata;
 	if(ret >= 0)
+#endif
 		ret = INT123_open_stream_handle(mh, iohandle);
 	return ret;
 }
-#endif
 
 int attribute_align_arg mpg123_open_feed(mpg123_handle *mh)
 {
