diff --git a/libs/musl/Makefile.in b/libs/musl/Makefile.in
index 16b0f1bf8574af2648366e1380b8bfa8cf366176..bcb14ffef7864d34cfaffe8810b755b4143edaf0 100644
--- a/libs/musl/Makefile.in
+++ b/libs/musl/Makefile.in
@@ -1,13 +1,8 @@
 EXTLIB    = libmusl.a
 EXTRAINCL = -I$(srcdir)/src/internal -I$(srcdir)/arch/generic
-EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE \
-            -fno-builtin
+EXTRADEFS = -D_ACRTIMP= -D_NO_CRT_MATH_INLINE
 
 SOURCES = \
-	src/complex/__cexp.c \
-	src/complex/carg.c \
-	src/complex/cargf.c \
-	src/complex/cexp.c \
 	src/math/__cos.c \
 	src/math/__cosdf.c \
 	src/math/__expo2.c \
