diff --git a/libs/png/pngrutil.c b/libs/png/pngrutil.c
index e7c7bbe48e9bafb888729d99b78f4275055faa57..d31dc21dae89e1e0c2970b0fabd1adea6ba97aab 100644
--- a/libs/png/pngrutil.c
+++ b/libs/png/pngrutil.c
@@ -1,6 +1,7 @@
+
 /* pngrutil.c - utilities to read a PNG file
  *
- * Copyright (c) 2018-2025 Cosmin Truta
+ * Copyright (c) 2018-2024 Cosmin Truta
  * Copyright (c) 1998-2002,2004,2006-2018 Glenn Randers-Pehrson
  * Copyright (c) 1996-1997 Andreas Dilger
  * Copyright (c) 1995-1996 Guy Eric Schalnat, Group 42, Inc.
@@ -17,26 +18,6 @@
 
 #ifdef PNG_READ_SUPPORTED
 
-/* The minimum 'zlib' stream is assumed to be just the 2 byte header, 5 bytes
- * minimum 'deflate' stream, and the 4 byte checksum.
- */
-#define LZ77Min  (2U+5U+4U)
-
-#ifdef PNG_READ_INTERLACING_SUPPORTED
-/* Arrays to facilitate interlacing - use pass (0 - 6) as index. */
-
-/* Start of interlace block */
-static const png_byte png_pass_start[7] = {0, 4, 0, 2, 0, 1, 0};
-/* Offset to next interlace block */
-static const png_byte png_pass_inc[7] = {8, 8, 4, 4, 2, 2, 1};
-/* Start of interlace block in the y direction */
-static const png_byte png_pass_ystart[7] = {0, 0, 4, 0, 2, 0, 1};
-/* Offset to next interlace block in the y direction */
-static const png_byte png_pass_yinc[7] = {8, 8, 8, 4, 4, 2, 2};
-
-/* TODO: Move these arrays to a common utility module to avoid duplication. */
-#endif
-
 png_uint_32 PNGAPI
 png_get_uint_31(png_const_structrp png_ptr, png_const_bytep buf)
 {
@@ -48,6 +29,30 @@ png_get_uint_31(png_const_structrp png_ptr, png_const_bytep buf)
    return uval;
 }
 
+#if defined(PNG_READ_gAMA_SUPPORTED) || defined(PNG_READ_cHRM_SUPPORTED)
+/* The following is a variation on the above for use with the fixed
+ * point values used for gAMA and cHRM.  Instead of png_error it
+ * issues a warning and returns (-1) - an invalid value because both
+ * gAMA and cHRM use *unsigned* integers for fixed point values.
+ */
+#define PNG_FIXED_ERROR (-1)
+
+static png_fixed_point /* PRIVATE */
+png_get_fixed_point(png_structrp png_ptr, png_const_bytep buf)
+{
+   png_uint_32 uval = png_get_uint_32(buf);
+
+   if (uval <= PNG_UINT_31_MAX)
+      return (png_fixed_point)uval; /* known to be in range */
+
+   /* The caller can turn off the warning by passing NULL. */
+   if (png_ptr != NULL)
+      png_warning(png_ptr, "PNG fixed point integer out of range");
+
+   return PNG_FIXED_ERROR;
+}
+#endif
+
 #ifdef PNG_READ_INT_FUNCTIONS_SUPPORTED
 /* NOTE: the read macros will obscure these definitions, so that if
  * PNG_USE_READ_MACROS is set the library will not use them internally,
@@ -144,38 +149,6 @@ png_read_sig(png_structrp png_ptr, png_inforp info_ptr)
       png_ptr->mode |= PNG_HAVE_PNG_SIGNATURE;
 }
 
-/* This function is called to verify that a chunk name is valid.
- * Do this using the bit-whacking approach from contrib/tools/pngfix.c
- *
- * Copied from libpng 1.7.
- */
-static int
-check_chunk_name(png_uint_32 name)
-{
-   png_uint_32 t;
-
-   /* Remove bit 5 from all but the reserved byte; this means
-    * every 8-bit unit must be in the range 65-90 to be valid.
-    * So bit 5 must be zero, bit 6 must be set and bit 7 zero.
-    */
-   name &= ~PNG_U32(32,32,0,32);
-   t = (name & ~0x1f1f1f1fU) ^ 0x40404040U;
-
-   /* Subtract 65 for each 8-bit quantity, this must not
-    * overflow and each byte must then be in the range 0-25.
-    */
-   name -= PNG_U32(65,65,65,65);
-   t |= name;
-
-   /* Subtract 26, handling the overflow which should set the
-    * top three bits of each byte.
-    */
-   name -= PNG_U32(25,25,25,26);
-   t |= ~name;
-
-   return (t & 0xe0e0e0e0U) == 0U;
-}
-
 /* Read the chunk header (length + type name).
  * Put the type name into png_ptr->chunk_name, and return the length.
  */
@@ -183,36 +156,33 @@ png_uint_32 /* PRIVATE */
 png_read_chunk_header(png_structrp png_ptr)
 {
    png_byte buf[8];
-   png_uint_32 chunk_name, length;
+   png_uint_32 length;
 
 #ifdef PNG_IO_STATE_SUPPORTED
    png_ptr->io_state = PNG_IO_READING | PNG_IO_CHUNK_HDR;
 #endif
 
-   /* Read the length and the chunk name.  png_struct::chunk_name is immediately
-    * updated even if they are detectably wrong.  This aids error message
-    * handling by allowing png_chunk_error to be used.
+   /* Read the length and the chunk name.
+    * This must be performed in a single I/O call.
     */
    png_read_data(png_ptr, buf, 8);
    length = png_get_uint_31(png_ptr, buf);
-   png_ptr->chunk_name = chunk_name = PNG_CHUNK_FROM_STRING(buf+4);
 
-   /* Reset the crc and run it over the chunk name. */
-   png_reset_crc(png_ptr);
-   png_calculate_crc(png_ptr, buf + 4, 4);
+   /* Put the chunk name into png_ptr->chunk_name. */
+   png_ptr->chunk_name = PNG_CHUNK_FROM_STRING(buf+4);
 
    png_debug2(0, "Reading chunk typeid = 0x%lx, length = %lu",
        (unsigned long)png_ptr->chunk_name, (unsigned long)length);
 
-   /* Sanity check the length (first by <= 0x80) and the chunk name.  An error
-    * here indicates a broken stream and libpng has no recovery from this.
-    */
-   if (buf[0] >= 0x80U)
-      png_chunk_error(png_ptr, "bad header (invalid length)");
+   /* Reset the crc and run it over the chunk name. */
+   png_reset_crc(png_ptr);
+   png_calculate_crc(png_ptr, buf + 4, 4);
 
    /* Check to see if chunk name is valid. */
-   if (!check_chunk_name(chunk_name))
-      png_chunk_error(png_ptr, "bad header (invalid type)");
+   png_check_chunk_name(png_ptr, png_ptr->chunk_name);
+
+   /* Check for too-large chunk length */
+   png_check_chunk_length(png_ptr, length);
 
 #ifdef PNG_IO_STATE_SUPPORTED
    png_ptr->io_state = PNG_IO_READING | PNG_IO_CHUNK_DATA;
@@ -232,85 +202,13 @@ png_crc_read(png_structrp png_ptr, png_bytep buf, png_uint_32 length)
    png_calculate_crc(png_ptr, buf, length);
 }
 
-/* Compare the CRC stored in the PNG file with that calculated by libpng from
- * the data it has read thus far.
- */
-static int
-png_crc_error(png_structrp png_ptr, int handle_as_ancillary)
-{
-   png_byte crc_bytes[4];
-   png_uint_32 crc;
-   int need_crc = 1;
-
-   /* There are four flags two for ancillary and two for critical chunks.  The
-    * default setting of these flags is all zero.
-    *
-    * PNG_FLAG_CRC_ANCILLARY_USE
-    * PNG_FLAG_CRC_ANCILLARY_NOWARN
-    *  USE+NOWARN: no CRC calculation (implemented here), else;
-    *  NOWARN:     png_chunk_error on error (implemented in png_crc_finish)
-    *  else:       png_chunk_warning on error (implemented in png_crc_finish)
-    *              This is the default.
-    *
-    *    I.e. NOWARN without USE produces png_chunk_error.  The default setting
-    *    where neither are set does the same thing.
-    *
-    * PNG_FLAG_CRC_CRITICAL_USE
-    * PNG_FLAG_CRC_CRITICAL_IGNORE
-    *  IGNORE: no CRC calculation (implemented here), else;
-    *  USE:    png_chunk_warning on error (implemented in png_crc_finish)
-    *  else:   png_chunk_error on error (implemented in png_crc_finish)
-    *          This is the default.
-    *
-    * This arose because of original mis-implementation and has persisted for
-    * compatibility reasons.
-    *
-    * TODO: the flag names are internal so maybe this can be changed to
-    * something comprehensible.
-    */
-   if (handle_as_ancillary || PNG_CHUNK_ANCILLARY(png_ptr->chunk_name) != 0)
-   {
-      if ((png_ptr->flags & PNG_FLAG_CRC_ANCILLARY_MASK) ==
-          (PNG_FLAG_CRC_ANCILLARY_USE | PNG_FLAG_CRC_ANCILLARY_NOWARN))
-         need_crc = 0;
-   }
-
-   else /* critical */
-   {
-      if ((png_ptr->flags & PNG_FLAG_CRC_CRITICAL_IGNORE) != 0)
-         need_crc = 0;
-   }
-
-#ifdef PNG_IO_STATE_SUPPORTED
-   png_ptr->io_state = PNG_IO_READING | PNG_IO_CHUNK_CRC;
-#endif
-
-   /* The chunk CRC must be serialized in a single I/O call. */
-   png_read_data(png_ptr, crc_bytes, 4);
-
-   if (need_crc != 0)
-   {
-      crc = png_get_uint_32(crc_bytes);
-      return crc != png_ptr->crc;
-   }
-
-   else
-      return 0;
-}
-
 /* Optionally skip data and then check the CRC.  Depending on whether we
  * are reading an ancillary or critical chunk, and how the program has set
  * things up, we may calculate the CRC on the data and print a message.
  * Returns '1' if there was a CRC error, '0' otherwise.
- *
- * There is one public version which is used in most places and another which
- * takes the value for the 'critical' flag to check.  This allows PLTE and IEND
- * handling code to ignore the CRC error and removes some confusing code
- * duplication.
  */
-static int
-png_crc_finish_critical(png_structrp png_ptr, png_uint_32 skip,
-      int handle_as_ancillary)
+int /* PRIVATE */
+png_crc_finish(png_structrp png_ptr, png_uint_32 skip)
 {
    /* The size of the local buffer for inflate is a good guess as to a
     * reasonable size to use for buffering reads from the application.
@@ -328,24 +226,14 @@ png_crc_finish_critical(png_structrp png_ptr, png_uint_32 skip,
       png_crc_read(png_ptr, tmpbuf, len);
    }
 
-   /* If 'handle_as_ancillary' has been requested and this is a critical chunk
-    * but PNG_FLAG_CRC_CRITICAL_IGNORE was set then png_read_crc did not, in
-    * fact, calculate the CRC so the ANCILLARY settings should not be used
-    * instead.
-    */
-   if (handle_as_ancillary &&
-       (png_ptr->flags & PNG_FLAG_CRC_CRITICAL_IGNORE) != 0)
-      handle_as_ancillary = 0;
-
-   /* TODO: this might be more comprehensible if png_crc_error was inlined here.
-    */
-   if (png_crc_error(png_ptr, handle_as_ancillary) != 0)
+   if (png_crc_error(png_ptr) != 0)
    {
-      /* See above for the explanation of how the flags work. */
-      if (handle_as_ancillary || PNG_CHUNK_ANCILLARY(png_ptr->chunk_name) != 0 ?
+      if (PNG_CHUNK_ANCILLARY(png_ptr->chunk_name) != 0 ?
           (png_ptr->flags & PNG_FLAG_CRC_ANCILLARY_NOWARN) == 0 :
           (png_ptr->flags & PNG_FLAG_CRC_CRITICAL_USE) != 0)
+      {
          png_chunk_warning(png_ptr, "CRC error");
+      }
 
       else
          png_chunk_error(png_ptr, "CRC error");
@@ -356,29 +244,61 @@ png_crc_finish_critical(png_structrp png_ptr, png_uint_32 skip,
    return 0;
 }
 
+/* Compare the CRC stored in the PNG file with that calculated by libpng from
+ * the data it has read thus far.
+ */
 int /* PRIVATE */
-png_crc_finish(png_structrp png_ptr, png_uint_32 skip)
+png_crc_error(png_structrp png_ptr)
 {
-   return png_crc_finish_critical(png_ptr, skip, 0/*critical handling*/);
+   png_byte crc_bytes[4];
+   png_uint_32 crc;
+   int need_crc = 1;
+
+   if (PNG_CHUNK_ANCILLARY(png_ptr->chunk_name) != 0)
+   {
+      if ((png_ptr->flags & PNG_FLAG_CRC_ANCILLARY_MASK) ==
+          (PNG_FLAG_CRC_ANCILLARY_USE | PNG_FLAG_CRC_ANCILLARY_NOWARN))
+         need_crc = 0;
+   }
+
+   else /* critical */
+   {
+      if ((png_ptr->flags & PNG_FLAG_CRC_CRITICAL_IGNORE) != 0)
+         need_crc = 0;
+   }
+
+#ifdef PNG_IO_STATE_SUPPORTED
+   png_ptr->io_state = PNG_IO_READING | PNG_IO_CHUNK_CRC;
+#endif
+
+   /* The chunk CRC must be serialized in a single I/O call. */
+   png_read_data(png_ptr, crc_bytes, 4);
+
+   if (need_crc != 0)
+   {
+      crc = png_get_uint_32(crc_bytes);
+      return crc != png_ptr->crc;
+   }
+
+   else
+      return 0;
 }
 
 #if defined(PNG_READ_iCCP_SUPPORTED) || defined(PNG_READ_iTXt_SUPPORTED) ||\
     defined(PNG_READ_pCAL_SUPPORTED) || defined(PNG_READ_sCAL_SUPPORTED) ||\
     defined(PNG_READ_sPLT_SUPPORTED) || defined(PNG_READ_tEXt_SUPPORTED) ||\
-    defined(PNG_READ_zTXt_SUPPORTED) || defined(PNG_READ_eXIf_SUPPORTED) ||\
-    defined(PNG_SEQUENTIAL_READ_SUPPORTED)
+    defined(PNG_READ_zTXt_SUPPORTED) || defined(PNG_SEQUENTIAL_READ_SUPPORTED)
 /* Manage the read buffer; this simply reallocates the buffer if it is not small
  * enough (or if it is not allocated).  The routine returns a pointer to the
  * buffer; if an error occurs and 'warn' is set the routine returns NULL, else
- * it will call png_error on failure.
+ * it will call png_error (via png_malloc) on failure.  (warn == 2 means
+ * 'silent').
  */
 static png_bytep
-png_read_buffer(png_structrp png_ptr, png_alloc_size_t new_size)
+png_read_buffer(png_structrp png_ptr, png_alloc_size_t new_size, int warn)
 {
    png_bytep buffer = png_ptr->read_buffer;
 
-   if (new_size > png_chunk_max(png_ptr)) return NULL;
-
    if (buffer != NULL && new_size > png_ptr->read_buffer_size)
    {
       png_ptr->read_buffer = NULL;
@@ -393,17 +313,24 @@ png_read_buffer(png_structrp png_ptr, png_alloc_size_t new_size)
 
       if (buffer != NULL)
       {
-#        ifndef PNG_NO_MEMZERO /* for detecting UIM bugs **only** */
-            memset(buffer, 0, new_size); /* just in case */
-#        endif
+         memset(buffer, 0, new_size); /* just in case */
          png_ptr->read_buffer = buffer;
          png_ptr->read_buffer_size = new_size;
       }
+
+      else if (warn < 2) /* else silent */
+      {
+         if (warn != 0)
+             png_chunk_warning(png_ptr, "insufficient memory to read chunk");
+
+         else
+             png_chunk_error(png_ptr, "insufficient memory to read chunk");
+      }
    }
 
    return buffer;
 }
-#endif /* READ_iCCP|iTXt|pCAL|sCAL|sPLT|tEXt|zTXt|eXIf|SEQUENTIAL_READ */
+#endif /* READ_iCCP|iTXt|pCAL|sCAL|sPLT|tEXt|zTXt|SEQUENTIAL_READ */
 
 /* png_inflate_claim: claim the zstream for some nefarious purpose that involves
  * decompression.  Returns Z_OK on success, else a zlib error code.  It checks
@@ -690,7 +617,16 @@ png_decompress_chunk(png_structrp png_ptr,
     * maybe a '\0' terminator too.  We have to assume that 'prefix_size' is
     * limited only by the maximum chunk size.
     */
-   png_alloc_size_t limit = png_chunk_max(png_ptr);
+   png_alloc_size_t limit = PNG_SIZE_MAX;
+
+# ifdef PNG_SET_USER_LIMITS_SUPPORTED
+   if (png_ptr->user_chunk_malloc_max > 0 &&
+       png_ptr->user_chunk_malloc_max < limit)
+      limit = png_ptr->user_chunk_malloc_max;
+# elif PNG_USER_CHUNK_MALLOC_MAX > 0
+   if (PNG_USER_CHUNK_MALLOC_MAX < limit)
+      limit = PNG_USER_CHUNK_MALLOC_MAX;
+# endif
 
    if (limit >= prefix_size + (terminate != 0))
    {
@@ -895,9 +831,9 @@ png_inflate_read(png_structrp png_ptr, png_bytep read_buffer, uInt read_size,
 }
 #endif /* READ_iCCP */
 
-/* CHUNK HANDLING */
 /* Read and check the IDHR chunk */
-static png_handle_result_code
+
+void /* PRIVATE */
 png_handle_IHDR(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_byte buf[13];
@@ -907,7 +843,12 @@ png_handle_IHDR(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    png_debug(1, "in png_handle_IHDR");
 
-   /* Length and position are checked by the caller. */
+   if ((png_ptr->mode & PNG_HAVE_IHDR) != 0)
+      png_chunk_error(png_ptr, "out of place");
+
+   /* Check the length */
+   if (length != 13)
+      png_chunk_error(png_ptr, "invalid");
 
    png_ptr->mode |= PNG_HAVE_IHDR;
 
@@ -961,196 +902,257 @@ png_handle_IHDR(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    png_debug1(3, "bit_depth = %d", png_ptr->bit_depth);
    png_debug1(3, "channels = %d", png_ptr->channels);
    png_debug1(3, "rowbytes = %lu", (unsigned long)png_ptr->rowbytes);
-
-   /* Rely on png_set_IHDR to completely validate the data and call png_error if
-    * it's wrong.
-    */
    png_set_IHDR(png_ptr, info_ptr, width, height, bit_depth,
        color_type, interlace_type, compression_type, filter_type);
-
-   return handled_ok;
-   PNG_UNUSED(length)
 }
 
 /* Read and check the palette */
-/* TODO: there are several obvious errors in this code when handling
- * out-of-place chunks and there is much over-complexity caused by trying to
- * patch up the problems.
- */
-static png_handle_result_code
+void /* PRIVATE */
 png_handle_PLTE(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
-   png_const_charp errmsg = NULL;
+   png_color palette[PNG_MAX_PALETTE_LENGTH];
+   int max_palette_length, num, i;
+#ifdef PNG_POINTER_INDEXING_SUPPORTED
+   png_colorp pal_ptr;
+#endif
 
    png_debug(1, "in png_handle_PLTE");
 
-   /* 1.6.47: consistency.  This used to be especially treated as a critical
-    * error even in an image which is not colour mapped, there isn't a good
-    * justification for treating some errors here one way and others another so
-    * everything uses the same logic.
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   /* Moved to before the 'after IDAT' check below because otherwise duplicate
+    * PLTE chunks are potentially ignored (the spec says there shall not be more
+    * than one PLTE, the error is not treated as benign, so this check trumps
+    * the requirement that PLTE appears before IDAT.)
     */
-   if ((png_ptr->mode & PNG_HAVE_PLTE) != 0)
-      errmsg = "duplicate";
+   else if ((png_ptr->mode & PNG_HAVE_PLTE) != 0)
+      png_chunk_error(png_ptr, "duplicate");
 
    else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
-      errmsg = "out of place";
+   {
+      /* This is benign because the non-benign error happened before, when an
+       * IDAT was encountered in a color-mapped image with no PLTE.
+       */
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
 
-   else if ((png_ptr->color_type & PNG_COLOR_MASK_COLOR) == 0)
-      errmsg = "ignored in grayscale PNG";
+   png_ptr->mode |= PNG_HAVE_PLTE;
 
-   else if (length > 3*PNG_MAX_PALETTE_LENGTH || (length % 3) != 0)
-      errmsg = "invalid";
+   if ((png_ptr->color_type & PNG_COLOR_MASK_COLOR) == 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "ignored in grayscale PNG");
+      return;
+   }
 
-   /* This drops PLTE in favour of tRNS or bKGD because both of those chunks
-    * can have an effect on the rendering of the image whereas PLTE only matters
-    * in the case of an 8-bit display with a decoder which controls the palette.
-    *
-    * The alternative here is to ignore the error and store the palette anyway;
-    * destroying the tRNS will definately cause problems.
-    *
-    * NOTE: the case of PNG_COLOR_TYPE_PALETTE need not be considered because
-    * the png_handle_ routines for the three 'after PLTE' chunks tRNS, bKGD and
-    * hIST all check for a preceding PLTE in these cases.
-    */
-   else if (png_ptr->color_type != PNG_COLOR_TYPE_PALETTE &&
-            (png_has_chunk(png_ptr, tRNS) || png_has_chunk(png_ptr, bKGD)))
-      errmsg = "out of place";
+#ifndef PNG_READ_OPT_PLTE_SUPPORTED
+   if (png_ptr->color_type != PNG_COLOR_TYPE_PALETTE)
+   {
+      png_crc_finish(png_ptr, length);
+      return;
+   }
+#endif
 
-   else
+   if (length > 3*PNG_MAX_PALETTE_LENGTH || length % 3)
    {
-      /* If the palette has 256 or fewer entries but is too large for the bit
-       * depth we don't issue an error to preserve the behavior of previous
-       * libpng versions. We silently truncate the unused extra palette entries
-       * here.
-       */
-      const unsigned max_palette_length =
-         (png_ptr->color_type == PNG_COLOR_TYPE_PALETTE) ?
-            1U << png_ptr->bit_depth : PNG_MAX_PALETTE_LENGTH;
+      png_crc_finish(png_ptr, length);
 
-      /* The cast is safe because 'length' is less than
-       * 3*PNG_MAX_PALETTE_LENGTH
-       */
-      const unsigned num = (length > 3U*max_palette_length) ?
-         max_palette_length : (unsigned)length / 3U;
+      if (png_ptr->color_type != PNG_COLOR_TYPE_PALETTE)
+         png_chunk_benign_error(png_ptr, "invalid");
 
-      unsigned i, j;
-      png_byte buf[3*PNG_MAX_PALETTE_LENGTH];
-      png_color palette[PNG_MAX_PALETTE_LENGTH];
+      else
+         png_chunk_error(png_ptr, "invalid");
 
-      /* Read the chunk into the buffer then read to the end of the chunk. */
-      png_crc_read(png_ptr, buf, num*3U);
-      png_crc_finish_critical(png_ptr, length - 3U*num,
-            /* Handle as ancillary if PLTE is optional: */
-            png_ptr->color_type != PNG_COLOR_TYPE_PALETTE);
+      return;
+   }
 
-      for (i = 0U, j = 0U; i < num; i++)
-      {
-         palette[i].red = buf[j++];
-         palette[i].green = buf[j++];
-         palette[i].blue = buf[j++];
-      }
+   /* The cast is safe because 'length' is less than 3*PNG_MAX_PALETTE_LENGTH */
+   num = (int)length / 3;
 
-      /* A valid PLTE chunk has been read */
-      png_ptr->mode |= PNG_HAVE_PLTE;
+   /* If the palette has 256 or fewer entries but is too large for the bit
+    * depth, we don't issue an error, to preserve the behavior of previous
+    * libpng versions. We silently truncate the unused extra palette entries
+    * here.
+    */
+   if (png_ptr->color_type == PNG_COLOR_TYPE_PALETTE)
+      max_palette_length = (1 << png_ptr->bit_depth);
+   else
+      max_palette_length = PNG_MAX_PALETTE_LENGTH;
 
-      /* TODO: png_set_PLTE has the side effect of setting png_ptr->palette to
-       * its own copy of the palette.  This has the side effect that when
-       * png_start_row is called (this happens after any call to
-       * png_read_update_info) the info_ptr palette gets changed.  This is
-       * extremely unexpected and confusing.
-       *
-       * REVIEW: there have been consistent bugs in the past about gamma and
-       * similar transforms to colour mapped images being useless because the
-       * modified palette cannot be accessed because of the above.
-       *
-       * CONSIDER: Fix this by not sharing the palette in this way.  But does
-       * this completely fix the problem?
-       */
-      png_set_PLTE(png_ptr, info_ptr, palette, num);
-      return handled_ok;
+   if (num > max_palette_length)
+      num = max_palette_length;
+
+#ifdef PNG_POINTER_INDEXING_SUPPORTED
+   for (i = 0, pal_ptr = palette; i < num; i++, pal_ptr++)
+   {
+      png_byte buf[3];
+
+      png_crc_read(png_ptr, buf, 3);
+      pal_ptr->red = buf[0];
+      pal_ptr->green = buf[1];
+      pal_ptr->blue = buf[2];
    }
+#else
+   for (i = 0; i < num; i++)
+   {
+      png_byte buf[3];
 
-   /* Here on error: errmsg is non NULL. */
+      png_crc_read(png_ptr, buf, 3);
+      /* Don't depend upon png_color being any order */
+      palette[i].red = buf[0];
+      palette[i].green = buf[1];
+      palette[i].blue = buf[2];
+   }
+#endif
+
+   /* If we actually need the PLTE chunk (ie for a paletted image), we do
+    * whatever the normal CRC configuration tells us.  However, if we
+    * have an RGB image, the PLTE can be considered ancillary, so
+    * we will act as though it is.
+    */
+#ifndef PNG_READ_OPT_PLTE_SUPPORTED
    if (png_ptr->color_type == PNG_COLOR_TYPE_PALETTE)
+#endif
    {
-      png_crc_finish(png_ptr, length);
-      png_chunk_error(png_ptr, errmsg);
+      png_crc_finish(png_ptr, (png_uint_32) (length - (unsigned int)num * 3));
    }
 
-   else /* not critical to this image */
+#ifndef PNG_READ_OPT_PLTE_SUPPORTED
+   else if (png_crc_error(png_ptr) != 0)  /* Only if we have a CRC error */
    {
-      png_crc_finish_critical(png_ptr, length, 1/*handle as ancillary*/);
-      png_chunk_benign_error(png_ptr, errmsg);
+      /* If we don't want to use the data from an ancillary chunk,
+       * we have two options: an error abort, or a warning and we
+       * ignore the data in this chunk (which should be OK, since
+       * it's considered ancillary for a RGB or RGBA image).
+       *
+       * IMPLEMENTATION NOTE: this is only here because png_crc_finish uses the
+       * chunk type to determine whether to check the ancillary or the critical
+       * flags.
+       */
+      if ((png_ptr->flags & PNG_FLAG_CRC_ANCILLARY_USE) == 0)
+      {
+         if ((png_ptr->flags & PNG_FLAG_CRC_ANCILLARY_NOWARN) != 0)
+            return;
+
+         else
+            png_chunk_error(png_ptr, "CRC error");
+      }
+
+      /* Otherwise, we (optionally) emit a warning and use the chunk. */
+      else if ((png_ptr->flags & PNG_FLAG_CRC_ANCILLARY_NOWARN) == 0)
+         png_chunk_warning(png_ptr, "CRC error");
    }
+#endif
 
-   /* Because PNG_UNUSED(errmsg) does not work if all the uses are compiled out
-    * (this does happen).
+   /* TODO: png_set_PLTE has the side effect of setting png_ptr->palette to its
+    * own copy of the palette.  This has the side effect that when png_start_row
+    * is called (this happens after any call to png_read_update_info) the
+    * info_ptr palette gets changed.  This is extremely unexpected and
+    * confusing.
+    *
+    * Fix this by not sharing the palette in this way.
     */
-   return errmsg != NULL ? handled_error : handled_error;
-}
+   png_set_PLTE(png_ptr, info_ptr, palette, num);
+
+   /* The three chunks, bKGD, hIST and tRNS *must* appear after PLTE and before
+    * IDAT.  Prior to 1.6.0 this was not checked; instead the code merely
+    * checked the apparent validity of a tRNS chunk inserted before PLTE on a
+    * palette PNG.  1.6.0 attempts to rigorously follow the standard and
+    * therefore does a benign error if the erroneous condition is detected *and*
+    * cancels the tRNS if the benign error returns.  The alternative is to
+    * amend the standard since it would be rather hypocritical of the standards
+    * maintainers to ignore it.
+    */
+#ifdef PNG_READ_tRNS_SUPPORTED
+   if (png_ptr->num_trans > 0 ||
+       (info_ptr != NULL && (info_ptr->valid & PNG_INFO_tRNS) != 0))
+   {
+      /* Cancel this because otherwise it would be used if the transforms
+       * require it.  Don't cancel the 'valid' flag because this would prevent
+       * detection of duplicate chunks.
+       */
+      png_ptr->num_trans = 0;
 
-/* On read the IDAT chunk is always handled specially, even if marked for
- * unknown handling (this is allowed), so:
- */
-#define png_handle_IDAT NULL
+      if (info_ptr != NULL)
+         info_ptr->num_trans = 0;
+
+      png_chunk_benign_error(png_ptr, "tRNS must be after");
+   }
+#endif
+
+#ifdef PNG_READ_hIST_SUPPORTED
+   if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_hIST) != 0)
+      png_chunk_benign_error(png_ptr, "hIST must be after");
+#endif
+
+#ifdef PNG_READ_bKGD_SUPPORTED
+   if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_bKGD) != 0)
+      png_chunk_benign_error(png_ptr, "bKGD must be after");
+#endif
+}
 
-static png_handle_result_code
+void /* PRIVATE */
 png_handle_IEND(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_debug(1, "in png_handle_IEND");
 
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0 ||
+       (png_ptr->mode & PNG_HAVE_IDAT) == 0)
+      png_chunk_error(png_ptr, "out of place");
+
    png_ptr->mode |= (PNG_AFTER_IDAT | PNG_HAVE_IEND);
 
+   png_crc_finish(png_ptr, length);
+
    if (length != 0)
       png_chunk_benign_error(png_ptr, "invalid");
 
-   png_crc_finish_critical(png_ptr, length, 1/*handle as ancillary*/);
-
-   return handled_ok;
    PNG_UNUSED(info_ptr)
 }
 
 #ifdef PNG_READ_gAMA_SUPPORTED
-static png_handle_result_code
+void /* PRIVATE */
 png_handle_gAMA(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
-   png_uint_32 ugamma;
+   png_fixed_point igamma;
    png_byte buf[4];
 
    png_debug(1, "in png_handle_gAMA");
 
-   png_crc_read(png_ptr, buf, 4);
-
-   if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
 
-   ugamma = png_get_uint_32(buf);
+   else if ((png_ptr->mode & (PNG_HAVE_IDAT|PNG_HAVE_PLTE)) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
 
-   if (ugamma > PNG_UINT_31_MAX)
+   if (length != 4)
    {
+      png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "invalid");
-      return handled_error;
+      return;
    }
 
-   png_set_gAMA_fixed(png_ptr, info_ptr, (png_fixed_point)/*SAFE*/ugamma);
+   png_crc_read(png_ptr, buf, 4);
 
-#ifdef PNG_READ_GAMMA_SUPPORTED
-      /* PNGv3: chunk precedence for gamma is cICP, [iCCP], sRGB, gAMA.  gAMA is
-       * at the end of the chain so simply check for an unset value.
-       */
-      if (png_ptr->chunk_gamma == 0)
-         png_ptr->chunk_gamma = (png_fixed_point)/*SAFE*/ugamma;
-#endif /*READ_GAMMA*/
+   if (png_crc_finish(png_ptr, 0) != 0)
+      return;
+
+   igamma = png_get_fixed_point(NULL, buf);
 
-   return handled_ok;
-   PNG_UNUSED(length)
+   png_colorspace_set_gamma(png_ptr, &png_ptr->colorspace, igamma);
+   png_colorspace_sync(png_ptr, info_ptr);
 }
-#else
-#  define png_handle_gAMA NULL
 #endif
 
 #ifdef PNG_READ_sBIT_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_sBIT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    unsigned int truelen, i;
@@ -1159,6 +1161,23 @@ png_handle_sBIT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    png_debug(1, "in png_handle_sBIT");
 
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & (PNG_HAVE_IDAT|PNG_HAVE_PLTE)) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_sBIT) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
+
    if (png_ptr->color_type == PNG_COLOR_TYPE_PALETTE)
    {
       truelen = 3;
@@ -1171,25 +1190,25 @@ png_handle_sBIT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       sample_depth = png_ptr->bit_depth;
    }
 
-   if (length != truelen)
+   if (length != truelen || length > 4)
    {
+      png_chunk_benign_error(png_ptr, "invalid");
       png_crc_finish(png_ptr, length);
-      png_chunk_benign_error(png_ptr, "bad length");
-      return handled_error;
+      return;
    }
 
    buf[0] = buf[1] = buf[2] = buf[3] = sample_depth;
    png_crc_read(png_ptr, buf, truelen);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    for (i=0; i<truelen; ++i)
    {
       if (buf[i] == 0 || buf[i] > sample_depth)
       {
          png_chunk_benign_error(png_ptr, "invalid");
-         return handled_error;
+         return;
       }
    }
 
@@ -1201,7 +1220,7 @@ png_handle_sBIT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       png_ptr->sig_bit.alpha = buf[3];
    }
 
-   else /* grayscale */
+   else
    {
       png_ptr->sig_bit.gray = buf[0];
       png_ptr->sig_bit.red = buf[0];
@@ -1211,132 +1230,133 @@ png_handle_sBIT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    }
 
    png_set_sBIT(png_ptr, info_ptr, &(png_ptr->sig_bit));
-   return handled_ok;
 }
-#else
-#  define png_handle_sBIT NULL
 #endif
 
 #ifdef PNG_READ_cHRM_SUPPORTED
-static png_int_32
-png_get_int_32_checked(png_const_bytep buf, int *error)
-{
-   png_uint_32 uval = png_get_uint_32(buf);
-   if ((uval & 0x80000000) == 0) /* non-negative */
-      return (png_int_32)uval;
-
-   uval = (uval ^ 0xffffffff) + 1;  /* 2's complement: -x = ~x+1 */
-   if ((uval & 0x80000000) == 0) /* no overflow */
-      return -(png_int_32)uval;
-
-   /* This version of png_get_int_32 has a way of returning the error to the
-    * caller, so:
-    */
-   *error = 1;
-   return 0; /* Safe */
-}
-
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_cHRM(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
-   int error = 0;
-   png_xy xy;
    png_byte buf[32];
+   png_xy xy;
 
    png_debug(1, "in png_handle_cHRM");
 
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & (PNG_HAVE_IDAT|PNG_HAVE_PLTE)) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   if (length != 32)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "invalid");
+      return;
+   }
+
    png_crc_read(png_ptr, buf, 32);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
-
-   xy.whitex = png_get_int_32_checked(buf +  0, &error);
-   xy.whitey = png_get_int_32_checked(buf +  4, &error);
-   xy.redx   = png_get_int_32_checked(buf +  8, &error);
-   xy.redy   = png_get_int_32_checked(buf + 12, &error);
-   xy.greenx = png_get_int_32_checked(buf + 16, &error);
-   xy.greeny = png_get_int_32_checked(buf + 20, &error);
-   xy.bluex  = png_get_int_32_checked(buf + 24, &error);
-   xy.bluey  = png_get_int_32_checked(buf + 28, &error);
+      return;
 
-   if (error)
+   xy.whitex = png_get_fixed_point(NULL, buf);
+   xy.whitey = png_get_fixed_point(NULL, buf + 4);
+   xy.redx   = png_get_fixed_point(NULL, buf + 8);
+   xy.redy   = png_get_fixed_point(NULL, buf + 12);
+   xy.greenx = png_get_fixed_point(NULL, buf + 16);
+   xy.greeny = png_get_fixed_point(NULL, buf + 20);
+   xy.bluex  = png_get_fixed_point(NULL, buf + 24);
+   xy.bluey  = png_get_fixed_point(NULL, buf + 28);
+
+   if (xy.whitex == PNG_FIXED_ERROR ||
+       xy.whitey == PNG_FIXED_ERROR ||
+       xy.redx   == PNG_FIXED_ERROR ||
+       xy.redy   == PNG_FIXED_ERROR ||
+       xy.greenx == PNG_FIXED_ERROR ||
+       xy.greeny == PNG_FIXED_ERROR ||
+       xy.bluex  == PNG_FIXED_ERROR ||
+       xy.bluey  == PNG_FIXED_ERROR)
    {
-      png_chunk_benign_error(png_ptr, "invalid");
-      return handled_error;
+      png_chunk_benign_error(png_ptr, "invalid values");
+      return;
    }
 
-   /* png_set_cHRM may complain about some of the values but this doesn't matter
-    * because it was a cHRM and it did have vaguely (if, perhaps, ridiculous)
-    * values.  Ridiculousity will be checked if the values are used later.
-    */
-   png_set_cHRM_fixed(png_ptr, info_ptr, xy.whitex, xy.whitey, xy.redx, xy.redy,
-         xy.greenx, xy.greeny, xy.bluex, xy.bluey);
+   /* If a colorspace error has already been output skip this chunk */
+   if ((png_ptr->colorspace.flags & PNG_COLORSPACE_INVALID) != 0)
+      return;
 
-   /* We only use 'chromaticities' for RGB to gray */
-#  ifdef PNG_READ_RGB_TO_GRAY_SUPPORTED
-      /* There is no need to check sRGB here, cICP is NYI and iCCP is not
-       * supported so just check mDCV.
-       */
-      if (!png_has_chunk(png_ptr, mDCV))
-      {
-         png_ptr->chromaticities = xy;
-      }
-#  endif /* READ_RGB_TO_GRAY */
+   if ((png_ptr->colorspace.flags & PNG_COLORSPACE_FROM_cHRM) != 0)
+   {
+      png_ptr->colorspace.flags |= PNG_COLORSPACE_INVALID;
+      png_colorspace_sync(png_ptr, info_ptr);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
 
-   return handled_ok;
-   PNG_UNUSED(length)
+   png_ptr->colorspace.flags |= PNG_COLORSPACE_FROM_cHRM;
+   (void)png_colorspace_set_chromaticities(png_ptr, &png_ptr->colorspace, &xy,
+       1/*prefer cHRM values*/);
+   png_colorspace_sync(png_ptr, info_ptr);
 }
-#else
-#  define png_handle_cHRM NULL
 #endif
 
 #ifdef PNG_READ_sRGB_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_sRGB(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_byte intent;
 
    png_debug(1, "in png_handle_sRGB");
 
-   png_crc_read(png_ptr, &intent, 1);
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
 
-   if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+   else if ((png_ptr->mode & (PNG_HAVE_IDAT|PNG_HAVE_PLTE)) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
 
-   /* This checks the range of the "rendering intent" because it is specified in
-    * the PNG spec itself; the "reserved" values will result in the chunk not
-    * being accepted, just as they do with the various "reserved" values in
-    * IHDR.
-    */
-   if (intent > 3/*PNGv3 spec*/)
+   if (length != 1)
    {
+      png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "invalid");
-      return handled_error;
+      return;
    }
 
-   png_set_sRGB(png_ptr, info_ptr, intent);
-   /* NOTE: png_struct::chromaticities is not set here because the RGB to gray
-    * coefficients are known without a need for the chromaticities.
-    */
+   png_crc_read(png_ptr, &intent, 1);
 
-#ifdef PNG_READ_GAMMA_SUPPORTED
-      /* PNGv3: chunk precedence for gamma is cICP, [iCCP], sRGB, gAMA.  iCCP is
-       * not supported by libpng so the only requirement is to check for cICP
-       * setting the gamma (this is NYI, but this check is safe.)
-       */
-      if (!png_has_chunk(png_ptr, cICP) || png_ptr->chunk_gamma == 0)
-         png_ptr->chunk_gamma = PNG_GAMMA_sRGB_INVERSE;
-#endif /*READ_GAMMA*/
+   if (png_crc_finish(png_ptr, 0) != 0)
+      return;
+
+   /* If a colorspace error has already been output skip this chunk */
+   if ((png_ptr->colorspace.flags & PNG_COLORSPACE_INVALID) != 0)
+      return;
+
+   /* Only one sRGB or iCCP chunk is allowed, use the HAVE_INTENT flag to detect
+    * this.
+    */
+   if ((png_ptr->colorspace.flags & PNG_COLORSPACE_HAVE_INTENT) != 0)
+   {
+      png_ptr->colorspace.flags |= PNG_COLORSPACE_INVALID;
+      png_colorspace_sync(png_ptr, info_ptr);
+      png_chunk_benign_error(png_ptr, "too many profiles");
+      return;
+   }
 
-   return handled_ok;
-   PNG_UNUSED(length)
+   (void)png_colorspace_set_sRGB(png_ptr, &png_ptr->colorspace, intent);
+   png_colorspace_sync(png_ptr, info_ptr);
 }
-#else
-#  define png_handle_sRGB NULL
 #endif /* READ_sRGB */
 
 #ifdef PNG_READ_iCCP_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 /* Note: this does not properly handle profiles that are > 64K under DOS */
 {
@@ -1345,10 +1365,44 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    png_debug(1, "in png_handle_iCCP");
 
-   /* PNGv3: allow PNG files with both sRGB and iCCP because the PNG spec only
-    * ever said that there "should" be only one, not "shall" and the PNGv3
-    * colour chunk precedence rules give a handling for this case anyway.
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & (PNG_HAVE_IDAT|PNG_HAVE_PLTE)) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   /* Consistent with all the above colorspace handling an obviously *invalid*
+    * chunk is just ignored, so does not invalidate the color space.  An
+    * alternative is to set the 'invalid' flags at the start of this routine
+    * and only clear them in they were not set before and all the tests pass.
     */
+
+   /* The keyword must be at least one character and there is a
+    * terminator (0) byte and the compression method byte, and the
+    * 'zlib' datastream is at least 11 bytes.
+    */
+   if (length < 14)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "too short");
+      return;
+   }
+
+   /* If a colorspace error has already been output skip this chunk */
+   if ((png_ptr->colorspace.flags & PNG_COLORSPACE_INVALID) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      return;
+   }
+
+   /* Only one sRGB or iCCP chunk is allowed, use the HAVE_INTENT flag to detect
+    * this.
+    */
+   if ((png_ptr->colorspace.flags & PNG_COLORSPACE_HAVE_INTENT) == 0)
    {
       uInt read_length, keyword_length;
       char keyword[81];
@@ -1358,16 +1412,19 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
        */
       read_length = 81; /* maximum */
       if (read_length > length)
-         read_length = (uInt)/*SAFE*/length;
+         read_length = (uInt)length;
 
       png_crc_read(png_ptr, (png_bytep)keyword, read_length);
       length -= read_length;
 
-      if (length < LZ77Min)
+      /* The minimum 'zlib' stream is assumed to be just the 2 byte header,
+       * 5 bytes minimum 'deflate' stream, and the 4 byte checksum.
+       */
+      if (length < 11)
       {
          png_crc_finish(png_ptr, length);
          png_chunk_benign_error(png_ptr, "too short");
-         return handled_error;
+         return;
       }
 
       keyword_length = 0;
@@ -1404,14 +1461,15 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
                    */
                   png_uint_32 profile_length = png_get_uint_32(profile_header);
 
-                  if (png_icc_check_length(png_ptr, keyword, profile_length) !=
-                      0)
+                  if (png_icc_check_length(png_ptr, &png_ptr->colorspace,
+                      keyword, profile_length) != 0)
                   {
                      /* The length is apparently ok, so we can check the 132
                       * byte header.
                       */
-                     if (png_icc_check_header(png_ptr, keyword, profile_length,
-                              profile_header, png_ptr->color_type) != 0)
+                     if (png_icc_check_header(png_ptr, &png_ptr->colorspace,
+                         keyword, profile_length, profile_header,
+                         png_ptr->color_type) != 0)
                      {
                         /* Now read the tag table; a variable size buffer is
                          * needed at this point, allocate one for the whole
@@ -1421,7 +1479,7 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
                         png_uint_32 tag_count =
                            png_get_uint_32(profile_header + 128);
                         png_bytep profile = png_read_buffer(png_ptr,
-                              profile_length);
+                            profile_length, 2/*silent*/);
 
                         if (profile != NULL)
                         {
@@ -1440,7 +1498,8 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
                            if (size == 0)
                            {
                               if (png_icc_check_tag_table(png_ptr,
-                                       keyword, profile_length, profile) != 0)
+                                  &png_ptr->colorspace, keyword, profile_length,
+                                  profile) != 0)
                               {
                                  /* The profile has been validated for basic
                                   * security issues, so read the whole thing in.
@@ -1472,6 +1531,13 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
                                     png_crc_finish(png_ptr, length);
                                     finished = 1;
 
+# if defined(PNG_sRGB_SUPPORTED) && PNG_sRGB_PROFILE_CHECKS >= 0
+                                    /* Check for a match against sRGB */
+                                    png_icc_set_sRGB(png_ptr,
+                                        &png_ptr->colorspace, profile,
+                                        png_ptr->zstream.adler);
+# endif
+
                                     /* Steal the profile for info_ptr. */
                                     if (info_ptr != NULL)
                                     {
@@ -1494,7 +1560,11 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
                                        }
 
                                        else
+                                       {
+                                          png_ptr->colorspace.flags |=
+                                             PNG_COLORSPACE_INVALID;
                                           errmsg = "out of memory";
+                                       }
                                     }
 
                                     /* else the profile remains in the read
@@ -1502,10 +1572,13 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
                                      * chunks.
                                      */
 
+                                    if (info_ptr != NULL)
+                                       png_colorspace_sync(png_ptr, info_ptr);
+
                                     if (errmsg == NULL)
                                     {
                                        png_ptr->zowner = 0;
-                                       return handled_ok;
+                                       return;
                                     }
                                  }
                                  if (errmsg == NULL)
@@ -1546,21 +1619,22 @@ png_handle_iCCP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
          errmsg = "bad keyword";
    }
 
+   else
+      errmsg = "too many profiles";
+
    /* Failure: the reason is in 'errmsg' */
    if (finished == 0)
       png_crc_finish(png_ptr, length);
 
+   png_ptr->colorspace.flags |= PNG_COLORSPACE_INVALID;
+   png_colorspace_sync(png_ptr, info_ptr);
    if (errmsg != NULL) /* else already output */
       png_chunk_benign_error(png_ptr, errmsg);
-
-   return handled_error;
 }
-#else
-#  define png_handle_iCCP NULL
 #endif /* READ_iCCP */
 
 #ifdef PNG_READ_sPLT_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 /* Note: this does not properly handle chunks that are > 64K under DOS */
 {
@@ -1581,24 +1655,43 @@ png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       if (png_ptr->user_chunk_cache_max == 1)
       {
          png_crc_finish(png_ptr, length);
-         return handled_error;
+         return;
       }
 
       if (--png_ptr->user_chunk_cache_max == 1)
       {
          png_warning(png_ptr, "No space in chunk cache for sPLT");
          png_crc_finish(png_ptr, length);
-         return handled_error;
+         return;
       }
    }
 #endif
 
-   buffer = png_read_buffer(png_ptr, length+1);
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+#ifdef PNG_MAX_MALLOC_64K
+   if (length > 65535U)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "too large to fit in memory");
+      return;
+   }
+#endif
+
+   buffer = png_read_buffer(png_ptr, length+1, 2/*silent*/);
    if (buffer == NULL)
    {
       png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "out of memory");
-      return handled_error;
+      return;
    }
 
 
@@ -1609,7 +1702,7 @@ png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    png_crc_read(png_ptr, buffer, length);
 
    if (png_crc_finish(png_ptr, skip) != 0)
-      return handled_error;
+      return;
 
    buffer[length] = 0;
 
@@ -1622,7 +1715,7 @@ png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    if (length < 2U || entry_start > buffer + (length - 2U))
    {
       png_warning(png_ptr, "malformed sPLT chunk");
-      return handled_error;
+      return;
    }
 
    new_palette.depth = *entry_start++;
@@ -1636,7 +1729,7 @@ png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    if ((data_length % (unsigned int)entry_size) != 0)
    {
       png_warning(png_ptr, "sPLT chunk has bad length");
-      return handled_error;
+      return;
    }
 
    dl = (png_uint_32)(data_length / (unsigned int)entry_size);
@@ -1645,7 +1738,7 @@ png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    if (dl > max_dl)
    {
       png_warning(png_ptr, "sPLT chunk too long");
-      return handled_error;
+      return;
    }
 
    new_palette.nentries = (png_int_32)(data_length / (unsigned int)entry_size);
@@ -1656,9 +1749,10 @@ png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    if (new_palette.entries == NULL)
    {
       png_warning(png_ptr, "sPLT chunk requires too much memory");
-      return handled_error;
+      return;
    }
 
+#ifdef PNG_POINTER_INDEXING_SUPPORTED
    for (i = 0; i < new_palette.nentries; i++)
    {
       pp = new_palette.entries + i;
@@ -1681,6 +1775,31 @@ png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
       pp->frequency = png_get_uint_16(entry_start); entry_start += 2;
    }
+#else
+   pp = new_palette.entries;
+
+   for (i = 0; i < new_palette.nentries; i++)
+   {
+
+      if (new_palette.depth == 8)
+      {
+         pp[i].red   = *entry_start++;
+         pp[i].green = *entry_start++;
+         pp[i].blue  = *entry_start++;
+         pp[i].alpha = *entry_start++;
+      }
+
+      else
+      {
+         pp[i].red   = png_get_uint_16(entry_start); entry_start += 2;
+         pp[i].green = png_get_uint_16(entry_start); entry_start += 2;
+         pp[i].blue  = png_get_uint_16(entry_start); entry_start += 2;
+         pp[i].alpha = png_get_uint_16(entry_start); entry_start += 2;
+      }
+
+      pp[i].frequency = png_get_uint_16(entry_start); entry_start += 2;
+   }
+#endif
 
    /* Discard all chunk data except the name and stash that */
    new_palette.name = (png_charp)buffer;
@@ -1688,20 +1807,34 @@ png_handle_sPLT(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    png_set_sPLT(png_ptr, info_ptr, &new_palette, 1);
 
    png_free(png_ptr, new_palette.entries);
-   return handled_ok;
 }
-#else
-#  define png_handle_sPLT NULL
 #endif /* READ_sPLT */
 
 #ifdef PNG_READ_tRNS_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_tRNS(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_byte readbuf[PNG_MAX_PALETTE_LENGTH];
 
    png_debug(1, "in png_handle_tRNS");
 
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   else if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_tRNS) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
+
    if (png_ptr->color_type == PNG_COLOR_TYPE_GRAY)
    {
       png_byte buf[2];
@@ -1710,7 +1843,7 @@ png_handle_tRNS(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       {
          png_crc_finish(png_ptr, length);
          png_chunk_benign_error(png_ptr, "invalid");
-         return handled_error;
+         return;
       }
 
       png_crc_read(png_ptr, buf, 2);
@@ -1726,7 +1859,7 @@ png_handle_tRNS(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       {
          png_crc_finish(png_ptr, length);
          png_chunk_benign_error(png_ptr, "invalid");
-         return handled_error;
+         return;
       }
 
       png_crc_read(png_ptr, buf, length);
@@ -1740,9 +1873,10 @@ png_handle_tRNS(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    {
       if ((png_ptr->mode & PNG_HAVE_PLTE) == 0)
       {
+         /* TODO: is this actually an error in the ISO spec? */
          png_crc_finish(png_ptr, length);
          png_chunk_benign_error(png_ptr, "out of place");
-         return handled_error;
+         return;
       }
 
       if (length > (unsigned int) png_ptr->num_palette ||
@@ -1751,7 +1885,7 @@ png_handle_tRNS(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       {
          png_crc_finish(png_ptr, length);
          png_chunk_benign_error(png_ptr, "invalid");
-         return handled_error;
+         return;
       }
 
       png_crc_read(png_ptr, readbuf, length);
@@ -1762,13 +1896,13 @@ png_handle_tRNS(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    {
       png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "invalid with alpha channel");
-      return handled_error;
+      return;
    }
 
    if (png_crc_finish(png_ptr, 0) != 0)
    {
       png_ptr->num_trans = 0;
-      return handled_error;
+      return;
    }
 
    /* TODO: this is a horrible side effect in the palette case because the
@@ -1777,14 +1911,11 @@ png_handle_tRNS(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
     */
    png_set_tRNS(png_ptr, info_ptr, readbuf, png_ptr->num_trans,
        &(png_ptr->trans_color));
-   return handled_ok;
 }
-#else
-#  define png_handle_tRNS NULL
 #endif
 
 #ifdef PNG_READ_bKGD_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_bKGD(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    unsigned int truelen;
@@ -1793,18 +1924,28 @@ png_handle_bKGD(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    png_debug(1, "in png_handle_bKGD");
 
-   if (png_ptr->color_type == PNG_COLOR_TYPE_PALETTE)
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0 ||
+       (png_ptr->color_type == PNG_COLOR_TYPE_PALETTE &&
+       (png_ptr->mode & PNG_HAVE_PLTE) == 0))
    {
-      if ((png_ptr->mode & PNG_HAVE_PLTE) == 0)
-      {
-         png_crc_finish(png_ptr, length);
-         png_chunk_benign_error(png_ptr, "out of place");
-         return handled_error;
-      }
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
 
-      truelen = 1;
+   else if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_bKGD) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
    }
 
+   if (png_ptr->color_type == PNG_COLOR_TYPE_PALETTE)
+      truelen = 1;
+
    else if ((png_ptr->color_type & PNG_COLOR_MASK_COLOR) != 0)
       truelen = 6;
 
@@ -1815,13 +1956,13 @@ png_handle_bKGD(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    {
       png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "invalid");
-      return handled_error;
+      return;
    }
 
    png_crc_read(png_ptr, buf, truelen);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    /* We convert the index value into RGB components so that we can allow
     * arbitrary RGB values for background when we have transparency, and
@@ -1837,7 +1978,7 @@ png_handle_bKGD(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
          if (buf[0] >= info_ptr->num_palette)
          {
             png_chunk_benign_error(png_ptr, "invalid index");
-            return handled_error;
+            return;
          }
 
          background.red = (png_uint_16)png_ptr->palette[buf[0]].red;
@@ -1858,7 +1999,7 @@ png_handle_bKGD(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
          if (buf[0] != 0 || buf[1] >= (unsigned int)(1 << png_ptr->bit_depth))
          {
             png_chunk_benign_error(png_ptr, "invalid gray level");
-            return handled_error;
+            return;
          }
       }
 
@@ -1876,7 +2017,7 @@ png_handle_bKGD(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
          if (buf[0] != 0 || buf[2] != 0 || buf[4] != 0)
          {
             png_chunk_benign_error(png_ptr, "invalid color");
-            return handled_error;
+            return;
          }
       }
 
@@ -1888,174 +2029,75 @@ png_handle_bKGD(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    }
 
    png_set_bKGD(png_ptr, info_ptr, &background);
-   return handled_ok;
 }
-#else
-#  define png_handle_bKGD NULL
-#endif
-
-#ifdef PNG_READ_cICP_SUPPORTED
-static png_handle_result_code /* PRIVATE */
-png_handle_cICP(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
-{
-   png_byte buf[4];
-
-   png_debug(1, "in png_handle_cICP");
-
-   png_crc_read(png_ptr, buf, 4);
-
-   if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
-
-   png_set_cICP(png_ptr, info_ptr, buf[0], buf[1],  buf[2], buf[3]);
-
-   /* We only use 'chromaticities' for RGB to gray */
-#  ifdef PNG_READ_RGB_TO_GRAY_SUPPORTED
-      if (!png_has_chunk(png_ptr, mDCV))
-      {
-         /* TODO: png_ptr->chromaticities = chromaticities; */
-      }
-#  endif /* READ_RGB_TO_GRAY */
-
-#ifdef PNG_READ_GAMMA_SUPPORTED
-      /* PNGv3: chunk precedence for gamma is cICP, [iCCP], sRGB, gAMA.  cICP is
-       * at the head so simply set the gamma if it can be determined.  If not
-       * chunk_gamma remains unchanged; sRGB and gAMA handling check it for
-       * being zero.
-       */
-      /* TODO: set png_struct::chunk_gamma when possible */
-#endif /*READ_GAMMA*/
-
-   return handled_ok;
-   PNG_UNUSED(length)
-}
-#else
-#  define png_handle_cICP NULL
-#endif
-
-#ifdef PNG_READ_cLLI_SUPPORTED
-static png_handle_result_code /* PRIVATE */
-png_handle_cLLI(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
-{
-   png_byte buf[8];
-
-   png_debug(1, "in png_handle_cLLI");
-
-   png_crc_read(png_ptr, buf, 8);
-
-   if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
-
-   /* The error checking happens here, this puts it in just one place: */
-   png_set_cLLI_fixed(png_ptr, info_ptr, png_get_uint_32(buf),
-         png_get_uint_32(buf+4));
-   return handled_ok;
-   PNG_UNUSED(length)
-}
-#else
-#  define png_handle_cLLI NULL
-#endif
-
-#ifdef PNG_READ_mDCV_SUPPORTED
-static png_handle_result_code /* PRIVATE */
-png_handle_mDCV(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
-{
-   png_xy chromaticities;
-   png_byte buf[24];
-
-   png_debug(1, "in png_handle_mDCV");
-
-   png_crc_read(png_ptr, buf, 24);
-
-   if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
-
-   /* The error checking happens here, this puts it in just one place.  The
-    * odd /50000 scaling factor makes it more difficult but the (x.y) values are
-    * only two bytes so a <<1 is safe.
-    *
-    * WARNING: the PNG specification defines the cHRM chunk to **start** with
-    * the white point (x,y).  The W3C PNG v3 specification puts the white point
-    * **after* R,G,B.  The x,y values in mDCV are also scaled by 50,000 and
-    * stored in just two bytes, whereas those in cHRM are scaled by 100,000 and
-    * stored in four bytes.  This is very, very confusing.  These APIs remove
-    * the confusion by copying the existing, well established, API.
-    */
-   chromaticities.redx   = png_get_uint_16(buf+ 0U) << 1; /* red x */
-   chromaticities.redy   = png_get_uint_16(buf+ 2U) << 1; /* red y */
-   chromaticities.greenx = png_get_uint_16(buf+ 4U) << 1; /* green x */
-   chromaticities.greeny = png_get_uint_16(buf+ 6U) << 1; /* green y */
-   chromaticities.bluex  = png_get_uint_16(buf+ 8U) << 1; /* blue x */
-   chromaticities.bluey  = png_get_uint_16(buf+10U) << 1; /* blue y */
-   chromaticities.whitex = png_get_uint_16(buf+12U) << 1; /* white x */
-   chromaticities.whitey = png_get_uint_16(buf+14U) << 1; /* white y */
-
-   png_set_mDCV_fixed(png_ptr, info_ptr,
-         chromaticities.whitex, chromaticities.whitey,
-         chromaticities.redx, chromaticities.redy,
-         chromaticities.greenx, chromaticities.greeny,
-         chromaticities.bluex, chromaticities.bluey,
-         png_get_uint_32(buf+16U), /* peak luminance */
-         png_get_uint_32(buf+20U));/* minimum perceivable luminance */
-
-   /* We only use 'chromaticities' for RGB to gray */
-#  ifdef PNG_READ_RGB_TO_GRAY_SUPPORTED
-      png_ptr->chromaticities = chromaticities;
-#  endif /* READ_RGB_TO_GRAY */
-
-   return handled_ok;
-   PNG_UNUSED(length)
-}
-#else
-#  define png_handle_mDCV NULL
 #endif
 
 #ifdef PNG_READ_eXIf_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_eXIf(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
-   png_bytep buffer = NULL;
+   unsigned int i;
 
    png_debug(1, "in png_handle_eXIf");
 
-   buffer = png_read_buffer(png_ptr, length);
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
 
-   if (buffer == NULL)
+   if (length < 2)
    {
       png_crc_finish(png_ptr, length);
-      png_chunk_benign_error(png_ptr, "out of memory");
-      return handled_error;
+      png_chunk_benign_error(png_ptr, "too short");
+      return;
    }
 
-   png_crc_read(png_ptr, buffer, length);
+   else if (info_ptr == NULL || (info_ptr->valid & PNG_INFO_eXIf) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
 
-   if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+   info_ptr->free_me |= PNG_FREE_EXIF;
 
-   /* PNGv3: the code used to check the byte order mark at the start for MM or
-    * II, however PNGv3 states that the the first 4 bytes should be checked.
-    * The caller ensures that there are four bytes available.
-    */
+   info_ptr->eXIf_buf = png_voidcast(png_bytep,
+             png_malloc_warn(png_ptr, length));
+
+   if (info_ptr->eXIf_buf == NULL)
    {
-      png_uint_32 header = png_get_uint_32(buffer);
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of memory");
+      return;
+   }
 
-      /* These numbers are copied from the PNGv3 spec: */
-      if (header != 0x49492A00 && header != 0x4D4D002A)
+   for (i = 0; i < length; i++)
+   {
+      png_byte buf[1];
+      png_crc_read(png_ptr, buf, 1);
+      info_ptr->eXIf_buf[i] = buf[0];
+      if (i == 1)
       {
-         png_chunk_benign_error(png_ptr, "invalid");
-         return handled_error;
+         if ((buf[0] != 'M' && buf[0] != 'I') ||
+             (info_ptr->eXIf_buf[0] != buf[0]))
+         {
+            png_crc_finish(png_ptr, length - 2);
+            png_chunk_benign_error(png_ptr, "incorrect byte-order specifier");
+            png_free(png_ptr, info_ptr->eXIf_buf);
+            info_ptr->eXIf_buf = NULL;
+            return;
+         }
       }
    }
 
-   png_set_eXIf_1(png_ptr, info_ptr, length, buffer);
-   return handled_ok;
+   if (png_crc_finish(png_ptr, 0) == 0)
+      png_set_eXIf_1(png_ptr, info_ptr, length, info_ptr->eXIf_buf);
+
+   png_free(png_ptr, info_ptr->eXIf_buf);
+   info_ptr->eXIf_buf = NULL;
 }
-#else
-#  define png_handle_eXIf NULL
 #endif
 
 #ifdef PNG_READ_hIST_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_hIST(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    unsigned int num, i;
@@ -2063,13 +2105,25 @@ png_handle_hIST(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    png_debug(1, "in png_handle_hIST");
 
-   /* This cast is safe because the chunk definition limits the length to a
-    * maximum of 1024 bytes.
-    *
-    * TODO: maybe use png_uint_32 anyway, not unsigned int, to reduce the
-    * casts.
-    */
-   num = (unsigned int)length / 2 ;
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0 ||
+       (png_ptr->mode & PNG_HAVE_PLTE) == 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   else if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_hIST) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
+
+   num = length / 2 ;
 
    if (length != num * 2 ||
        num != (unsigned int)png_ptr->num_palette ||
@@ -2077,7 +2131,7 @@ png_handle_hIST(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    {
       png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "invalid");
-      return handled_error;
+      return;
    }
 
    for (i = 0; i < num; i++)
@@ -2089,17 +2143,14 @@ png_handle_hIST(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    }
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    png_set_hIST(png_ptr, info_ptr, readbuf);
-   return handled_ok;
 }
-#else
-#  define png_handle_hIST NULL
 #endif
 
 #ifdef PNG_READ_pHYs_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_pHYs(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_byte buf[9];
@@ -2108,24 +2159,44 @@ png_handle_pHYs(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    png_debug(1, "in png_handle_pHYs");
 
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   else if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_pHYs) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
+
+   if (length != 9)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "invalid");
+      return;
+   }
+
    png_crc_read(png_ptr, buf, 9);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    res_x = png_get_uint_32(buf);
    res_y = png_get_uint_32(buf + 4);
    unit_type = buf[8];
    png_set_pHYs(png_ptr, info_ptr, res_x, res_y, unit_type);
-   return handled_ok;
-   PNG_UNUSED(length)
 }
-#else
-#  define png_handle_pHYs NULL
 #endif
 
 #ifdef PNG_READ_oFFs_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_oFFs(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_byte buf[9];
@@ -2134,25 +2205,45 @@ png_handle_oFFs(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    png_debug(1, "in png_handle_oFFs");
 
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   else if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_oFFs) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
+
+   if (length != 9)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "invalid");
+      return;
+   }
+
    png_crc_read(png_ptr, buf, 9);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    offset_x = png_get_int_32(buf);
    offset_y = png_get_int_32(buf + 4);
    unit_type = buf[8];
    png_set_oFFs(png_ptr, info_ptr, offset_x, offset_y, unit_type);
-   return handled_ok;
-   PNG_UNUSED(length)
 }
-#else
-#  define png_handle_oFFs NULL
 #endif
 
 #ifdef PNG_READ_pCAL_SUPPORTED
 /* Read the pCAL chunk (described in the PNG Extensions document) */
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_pCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_int_32 X0, X1;
@@ -2162,22 +2253,40 @@ png_handle_pCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    int i;
 
    png_debug(1, "in png_handle_pCAL");
+
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   else if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_pCAL) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
+
    png_debug1(2, "Allocating and reading pCAL chunk data (%u bytes)",
        length + 1);
 
-   buffer = png_read_buffer(png_ptr, length+1);
+   buffer = png_read_buffer(png_ptr, length+1, 2/*silent*/);
 
    if (buffer == NULL)
    {
       png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "out of memory");
-      return handled_error;
+      return;
    }
 
    png_crc_read(png_ptr, buffer, length);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    buffer[length] = 0; /* Null terminate the last string */
 
@@ -2193,7 +2302,7 @@ png_handle_pCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    if (endptr - buf <= 12)
    {
       png_chunk_benign_error(png_ptr, "invalid");
-      return handled_error;
+      return;
    }
 
    png_debug(3, "Reading pCAL X0, X1, type, nparams, and units");
@@ -2213,7 +2322,7 @@ png_handle_pCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
        (type == PNG_EQUATION_HYPERBOLIC && nparams != 4))
    {
       png_chunk_benign_error(png_ptr, "invalid parameter count");
-      return handled_error;
+      return;
    }
 
    else if (type >= PNG_EQUATION_LAST)
@@ -2232,7 +2341,7 @@ png_handle_pCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    if (params == NULL)
    {
       png_chunk_benign_error(png_ptr, "out of memory");
-      return handled_error;
+      return;
    }
 
    /* Get pointers to the start of each parameter string. */
@@ -2250,29 +2359,20 @@ png_handle_pCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       {
          png_free(png_ptr, params);
          png_chunk_benign_error(png_ptr, "invalid data");
-         return handled_error;
+         return;
       }
    }
 
    png_set_pCAL(png_ptr, info_ptr, (png_charp)buffer, X0, X1, type, nparams,
        (png_charp)units, params);
 
-   /* TODO: BUG: png_set_pCAL calls png_chunk_report which, in this case, calls
-    * png_benign_error and that can error out.
-    *
-    * png_read_buffer needs to be allocated with space for both nparams and the
-    * parameter strings.  Not hard to do.
-    */
    png_free(png_ptr, params);
-   return handled_ok;
 }
-#else
-#  define png_handle_pCAL NULL
 #endif
 
 #ifdef PNG_READ_sCAL_SUPPORTED
 /* Read the sCAL chunk */
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_sCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_bytep buffer;
@@ -2280,29 +2380,55 @@ png_handle_sCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    int state;
 
    png_debug(1, "in png_handle_sCAL");
+
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "out of place");
+      return;
+   }
+
+   else if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_sCAL) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
+
+   /* Need unit type, width, \0, height: minimum 4 bytes */
+   else if (length < 4)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "invalid");
+      return;
+   }
+
    png_debug1(2, "Allocating and reading sCAL chunk data (%u bytes)",
        length + 1);
 
-   buffer = png_read_buffer(png_ptr, length+1);
+   buffer = png_read_buffer(png_ptr, length+1, 2/*silent*/);
 
    if (buffer == NULL)
    {
-      png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "out of memory");
-      return handled_error;
+      png_crc_finish(png_ptr, length);
+      return;
    }
 
    png_crc_read(png_ptr, buffer, length);
    buffer[length] = 0; /* Null terminate the last string */
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    /* Validate the unit. */
    if (buffer[0] != 1 && buffer[0] != 2)
    {
       png_chunk_benign_error(png_ptr, "invalid unit");
-      return handled_error;
+      return;
    }
 
    /* Validate the ASCII numbers, need two ASCII numbers separated by
@@ -2331,22 +2457,15 @@ png_handle_sCAL(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
          png_chunk_benign_error(png_ptr, "non-positive height");
 
       else
-      {
          /* This is the (only) success case. */
          png_set_sCAL_s(png_ptr, info_ptr, buffer[0],
              (png_charp)buffer+1, (png_charp)buffer+heighti);
-         return handled_ok;
-      }
    }
-
-   return handled_error;
 }
-#else
-#  define png_handle_sCAL NULL
 #endif
 
 #ifdef PNG_READ_tIME_SUPPORTED
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_tIME(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_byte buf[7];
@@ -2354,17 +2473,30 @@ png_handle_tIME(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    png_debug(1, "in png_handle_tIME");
 
-   /* TODO: what is this doing here?  It should be happened in pngread.c and
-    * pngpread.c, although it could be moved to png_handle_chunk below and
-    * thereby avoid some code duplication.
-    */
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   else if (info_ptr != NULL && (info_ptr->valid & PNG_INFO_tIME) != 0)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "duplicate");
+      return;
+   }
+
    if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
       png_ptr->mode |= PNG_AFTER_IDAT;
 
+   if (length != 7)
+   {
+      png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "invalid");
+      return;
+   }
+
    png_crc_read(png_ptr, buf, 7);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    mod_time.second = buf[6];
    mod_time.minute = buf[5];
@@ -2374,16 +2506,12 @@ png_handle_tIME(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    mod_time.year = png_get_uint_16(buf);
 
    png_set_tIME(png_ptr, info_ptr, &mod_time);
-   return handled_ok;
-   PNG_UNUSED(length)
 }
-#else
-#  define png_handle_tIME NULL
 #endif
 
 #ifdef PNG_READ_tEXt_SUPPORTED
 /* Note: this does not properly handle chunks that are > 64K under DOS */
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_tEXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_text  text_info;
@@ -2400,31 +2528,45 @@ png_handle_tEXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       if (png_ptr->user_chunk_cache_max == 1)
       {
          png_crc_finish(png_ptr, length);
-         return handled_error;
+         return;
       }
 
       if (--png_ptr->user_chunk_cache_max == 1)
       {
          png_crc_finish(png_ptr, length);
          png_chunk_benign_error(png_ptr, "no space in chunk cache");
-         return handled_error;
+         return;
       }
    }
 #endif
 
-   buffer = png_read_buffer(png_ptr, length+1);
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
 
-   if (buffer == NULL)
+   if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+      png_ptr->mode |= PNG_AFTER_IDAT;
+
+#ifdef PNG_MAX_MALLOC_64K
+   if (length > 65535U)
    {
       png_crc_finish(png_ptr, length);
+      png_chunk_benign_error(png_ptr, "too large to fit in memory");
+      return;
+   }
+#endif
+
+   buffer = png_read_buffer(png_ptr, length+1, 1/*warn*/);
+
+   if (buffer == NULL)
+   {
       png_chunk_benign_error(png_ptr, "out of memory");
-      return handled_error;
+      return;
    }
 
    png_crc_read(png_ptr, buffer, length);
 
    if (png_crc_finish(png_ptr, skip) != 0)
-      return handled_error;
+      return;
 
    key = (png_charp)buffer;
    key[length] = 0;
@@ -2443,19 +2585,14 @@ png_handle_tEXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
    text_info.text = text;
    text_info.text_length = strlen(text);
 
-   if (png_set_text_2(png_ptr, info_ptr, &text_info, 1) == 0)
-      return handled_ok;
-
-   png_chunk_benign_error(png_ptr, "out of memory");
-   return handled_error;
+   if (png_set_text_2(png_ptr, info_ptr, &text_info, 1) != 0)
+      png_warning(png_ptr, "Insufficient memory to process text chunk");
 }
-#else
-#  define png_handle_tEXt NULL
 #endif
 
 #ifdef PNG_READ_zTXt_SUPPORTED
 /* Note: this does not correctly handle chunks that are > 64K under DOS */
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_zTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_const_charp errmsg = NULL;
@@ -2470,35 +2607,40 @@ png_handle_zTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       if (png_ptr->user_chunk_cache_max == 1)
       {
          png_crc_finish(png_ptr, length);
-         return handled_error;
+         return;
       }
 
       if (--png_ptr->user_chunk_cache_max == 1)
       {
          png_crc_finish(png_ptr, length);
          png_chunk_benign_error(png_ptr, "no space in chunk cache");
-         return handled_error;
+         return;
       }
    }
 #endif
 
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+      png_ptr->mode |= PNG_AFTER_IDAT;
+
    /* Note, "length" is sufficient here; we won't be adding
-    * a null terminator later.  The limit check in png_handle_chunk should be
-    * sufficient.
+    * a null terminator later.
     */
-   buffer = png_read_buffer(png_ptr, length);
+   buffer = png_read_buffer(png_ptr, length, 2/*silent*/);
 
    if (buffer == NULL)
    {
       png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "out of memory");
-      return handled_error;
+      return;
    }
 
    png_crc_read(png_ptr, buffer, length);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    /* TODO: also check that the keyword contents match the spec! */
    for (keyword_length = 0;
@@ -2551,10 +2693,8 @@ png_handle_zTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
             text.lang = NULL;
             text.lang_key = NULL;
 
-            if (png_set_text_2(png_ptr, info_ptr, &text, 1) == 0)
-               return handled_ok;
-
-            errmsg = "out of memory";
+            if (png_set_text_2(png_ptr, info_ptr, &text, 1) != 0)
+               errmsg = "insufficient memory";
          }
       }
 
@@ -2562,16 +2702,14 @@ png_handle_zTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
          errmsg = png_ptr->zstream.msg;
    }
 
-   png_chunk_benign_error(png_ptr, errmsg);
-   return handled_error;
+   if (errmsg != NULL)
+      png_chunk_benign_error(png_ptr, errmsg);
 }
-#else
-#  define png_handle_zTXt NULL
 #endif
 
 #ifdef PNG_READ_iTXt_SUPPORTED
 /* Note: this does not correctly handle chunks that are > 64K under DOS */
-static png_handle_result_code /* PRIVATE */
+void /* PRIVATE */
 png_handle_iTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 {
    png_const_charp errmsg = NULL;
@@ -2586,31 +2724,37 @@ png_handle_iTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
       if (png_ptr->user_chunk_cache_max == 1)
       {
          png_crc_finish(png_ptr, length);
-         return handled_error;
+         return;
       }
 
       if (--png_ptr->user_chunk_cache_max == 1)
       {
          png_crc_finish(png_ptr, length);
          png_chunk_benign_error(png_ptr, "no space in chunk cache");
-         return handled_error;
+         return;
       }
    }
 #endif
 
-   buffer = png_read_buffer(png_ptr, length+1);
+   if ((png_ptr->mode & PNG_HAVE_IHDR) == 0)
+      png_chunk_error(png_ptr, "missing IHDR");
+
+   if ((png_ptr->mode & PNG_HAVE_IDAT) != 0)
+      png_ptr->mode |= PNG_AFTER_IDAT;
+
+   buffer = png_read_buffer(png_ptr, length+1, 1/*warn*/);
 
    if (buffer == NULL)
    {
       png_crc_finish(png_ptr, length);
       png_chunk_benign_error(png_ptr, "out of memory");
-      return handled_error;
+      return;
    }
 
    png_crc_read(png_ptr, buffer, length);
 
    if (png_crc_finish(png_ptr, 0) != 0)
-      return handled_error;
+      return;
 
    /* First the keyword. */
    for (prefix_length=0;
@@ -2700,10 +2844,8 @@ png_handle_iTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
          text.text_length = 0;
          text.itxt_length = uncompressed_length;
 
-         if (png_set_text_2(png_ptr, info_ptr, &text, 1) == 0)
-            return handled_ok;
-
-         errmsg = "out of memory";
+         if (png_set_text_2(png_ptr, info_ptr, &text, 1) != 0)
+            errmsg = "insufficient memory";
       }
    }
 
@@ -2712,10 +2854,7 @@ png_handle_iTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 
    if (errmsg != NULL)
       png_chunk_benign_error(png_ptr, errmsg);
-   return handled_error;
 }
-#else
-#  define png_handle_iTXt NULL
 #endif
 
 #ifdef PNG_READ_UNKNOWN_CHUNKS_SUPPORTED
@@ -2723,7 +2862,7 @@ png_handle_iTXt(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
 static int
 png_cache_unknown_chunk(png_structrp png_ptr, png_uint_32 length)
 {
-   const png_alloc_size_t limit = png_chunk_max(png_ptr);
+   png_alloc_size_t limit = PNG_SIZE_MAX;
 
    if (png_ptr->unknown_chunk.data != NULL)
    {
@@ -2731,6 +2870,16 @@ png_cache_unknown_chunk(png_structrp png_ptr, png_uint_32 length)
       png_ptr->unknown_chunk.data = NULL;
    }
 
+#  ifdef PNG_SET_USER_LIMITS_SUPPORTED
+   if (png_ptr->user_chunk_malloc_max > 0 &&
+       png_ptr->user_chunk_malloc_max < limit)
+      limit = png_ptr->user_chunk_malloc_max;
+
+#  elif PNG_USER_CHUNK_MALLOC_MAX > 0
+   if (PNG_USER_CHUNK_MALLOC_MAX < limit)
+      limit = PNG_USER_CHUNK_MALLOC_MAX;
+#  endif
+
    if (length <= limit)
    {
       PNG_CSTRING_FROM_CHUNK(png_ptr->unknown_chunk.name, png_ptr->chunk_name);
@@ -2769,11 +2918,11 @@ png_cache_unknown_chunk(png_structrp png_ptr, png_uint_32 length)
 #endif /* READ_UNKNOWN_CHUNKS */
 
 /* Handle an unknown, or known but disabled, chunk */
-png_handle_result_code /*PRIVATE*/
+void /* PRIVATE */
 png_handle_unknown(png_structrp png_ptr, png_inforp info_ptr,
     png_uint_32 length, int keep)
 {
-   png_handle_result_code handled = handled_discarded; /* the default */
+   int handled = 0; /* the chunk was handled */
 
    png_debug(1, "in png_handle_unknown");
 
@@ -2820,7 +2969,7 @@ png_handle_unknown(png_structrp png_ptr, png_inforp info_ptr,
           *           error at this point unless it is to be saved.
           * positive: The chunk was handled, libpng will ignore/discard it.
           */
-         if (ret < 0) /* handled_error */
+         if (ret < 0)
             png_chunk_error(png_ptr, "error in user chunk");
 
          else if (ret == 0)
@@ -2854,7 +3003,7 @@ png_handle_unknown(png_structrp png_ptr, png_inforp info_ptr,
 
          else /* chunk was handled */
          {
-            handled = handled_ok;
+            handled = 1;
             /* Critical chunks can be safely discarded at this point. */
             keep = PNG_HANDLE_CHUNK_NEVER;
          }
@@ -2939,7 +3088,7 @@ png_handle_unknown(png_structrp png_ptr, png_inforp info_ptr,
              */
             png_set_unknown_chunks(png_ptr, info_ptr,
                 &png_ptr->unknown_chunk, 1);
-            handled = handled_saved;
+            handled = 1;
 #  ifdef PNG_USER_LIMITS_SUPPORTED
             break;
       }
@@ -2965,267 +3114,79 @@ png_handle_unknown(png_structrp png_ptr, png_inforp info_ptr,
 #endif /* !READ_UNKNOWN_CHUNKS */
 
    /* Check for unhandled critical chunks */
-   if (handled < handled_saved && PNG_CHUNK_CRITICAL(png_ptr->chunk_name))
+   if (handled == 0 && PNG_CHUNK_CRITICAL(png_ptr->chunk_name))
       png_chunk_error(png_ptr, "unhandled critical chunk");
-
-   return handled;
 }
 
-/* APNG handling: the minimal implementation of APNG handling in libpng 1.6
- * requires that those significant applications which already handle APNG not
- * get hosed.  To do this ensure the code here will have to ensure than APNG
- * data by default (at least in 1.6) gets stored in the unknown chunk list.
- * Maybe this can be relaxed in a few years but at present it's just the only
- * safe way.
- *
- * ATM just cause unknown handling for all three chunks:
+/* This function is called to verify that a chunk name is valid.
+ * This function can't have the "critical chunk check" incorporated
+ * into it, since in the future we will need to be able to call user
+ * functions to handle unknown critical chunks after we check that
+ * the chunk name itself is valid.
  */
-#define png_handle_acTL NULL
-#define png_handle_fcTL NULL
-#define png_handle_fdAT NULL
 
-/*
- * 1.6.47: This is the new table driven interface to all the chunk handling.
+/* Bit hacking: the test for an invalid byte in the 4 byte chunk name is:
  *
- * The table describes the PNG standard rules for **reading** known chunks -
- * every chunk which has an entry in PNG_KNOWN_CHUNKS.  The table contains an
- * entry for each PNG_INDEX_cHNK describing the rules.
- *
- * In this initial version the only information in the entry is the
- * png_handle_cHNK function for the chunk in question.  When chunk support is
- * compiled out the entry will be NULL.
+ * ((c) < 65 || (c) > 122 || ((c) > 90 && (c) < 97))
  */
-static const struct
-{
-   png_handle_result_code (*handler)(
-         png_structrp, png_inforp, png_uint_32 length);
-      /* A chunk-specific 'handler', NULL if the chunk is not supported in this
-       * build.
-       */
-
-   /* Crushing these values helps on modern 32-bit architectures because the
-    * pointer and the following bit fields both end up requiring 32 bits.
-    * Typically this will halve the table size.  On 64-bit architectures the
-    * table entries will typically be 8 bytes.
-    */
-   png_uint_32 max_length :12; /* Length min, max in bytes */
-   png_uint_32 min_length :8;
-      /* Length errors on critical chunks have special handling to preserve the
-       * existing behaviour in libpng 1.6.  Anciallary chunks are checked below
-       * and produce a 'benign' error.
-       */
-   png_uint_32 pos_before :4; /* PNG_HAVE_ values chunk must precede */
-   png_uint_32 pos_after  :4; /* PNG_HAVE_ values chunk must follow */
-      /* NOTE: PLTE, tRNS and bKGD require special handling which depends on
-       * the colour type of the base image.
-       */
-   png_uint_32 multiple   :1; /* Multiple occurences permitted */
-      /* This is enabled for PLTE because PLTE may, in practice, be optional */
-}
-read_chunks[PNG_INDEX_unknown] =
-{
-   /* Definitions as above but done indirectly by #define so that
-    * PNG_KNOWN_CHUNKS can be used safely to build the table in order.
-    *
-    * Each CDcHNK definition lists the values for the parameters **after**
-    * the first, 'handler', function.  'handler' is NULL when the chunk has no
-    * compiled in support.
-    */
-#  define NoCheck 0x801U      /* Do not check the maximum length */
-#  define Limit   0x802U      /* Limit to png_chunk_max bytes */
-#  define LKMin   3U+LZ77Min  /* Minimum length of keyword+LZ77 */
-
-#define hIHDR PNG_HAVE_IHDR
-#define hPLTE PNG_HAVE_PLTE
-#define hIDAT PNG_HAVE_IDAT
-   /* For the two chunks, tRNS and bKGD which can occur in PNGs without a PLTE
-    * but must occur after the PLTE use this and put the check in the handler
-    * routine for colour mapped images were PLTE is required.  Also put a check
-    * in PLTE for other image types to drop the PLTE if tRNS or bKGD have been
-    * seen.
-    */
-#define hCOL  (PNG_HAVE_PLTE|PNG_HAVE_IDAT)
-   /* Used for the decoding chunks which must be before PLTE. */
-#define aIDAT PNG_AFTER_IDAT
-
-   /* Chunks from W3C PNG v3: */
-   /*       cHNK  max_len,   min, before, after, multiple */
-#  define CDIHDR      13U,   13U,  hIHDR,     0,        0
-#  define CDPLTE  NoCheck,    0U,      0, hIHDR,        1
-      /* PLTE errors are only critical for colour-map images, consequently the
-       * hander does all the checks.
-       */
-#  define CDIDAT  NoCheck,    0U,  aIDAT, hIHDR,        1
-#  define CDIEND  NoCheck,    0U,      0, aIDAT,        0
-      /* Historically data was allowed in IEND */
-#  define CDtRNS     256U,    0U,  hIDAT, hIHDR,        0
-#  define CDcHRM      32U,   32U,   hCOL, hIHDR,        0
-#  define CDgAMA       4U,    4U,   hCOL, hIHDR,        0
-#  define CDiCCP  NoCheck, LKMin,   hCOL, hIHDR,        0
-#  define CDsBIT       4U,    1U,   hCOL, hIHDR,        0
-#  define CDsRGB       1U,    1U,   hCOL, hIHDR,        0
-#  define CDcICP       4U,    4U,   hCOL, hIHDR,        0
-#  define CDmDCV      24U,   24U,   hCOL, hIHDR,        0
-#  define CDeXIf    Limit,    4U,      0, hIHDR,        0
-#  define CDcLLI       8U,    8U,   hCOL, hIHDR,        0
-#  define CDtEXt  NoCheck,    2U,      0, hIHDR,        1
-      /* Allocates 'length+1'; checked in the handler */
-#  define CDzTXt    Limit, LKMin,      0, hIHDR,        1
-#  define CDiTXt  NoCheck,    6U,      0, hIHDR,        1
-      /* Allocates 'length+1'; checked in the handler */
-#  define CDbKGD       6U,    1U,  hIDAT, hIHDR,        0
-#  define CDhIST    1024U,    0U,  hPLTE, hIHDR,        0
-#  define CDpHYs       9U,    9U,  hIDAT, hIHDR,        0
-#  define CDsPLT  NoCheck,    3U,  hIDAT, hIHDR,        1
-      /* Allocates 'length+1'; checked in the handler */
-#  define CDtIME       7U,    7U,      0, hIHDR,        0
-#  define CDacTL       8U,    8U,  hIDAT, hIHDR,        0
-#  define CDfcTL      25U,   26U,      0, hIHDR,        1
-#  define CDfdAT    Limit,    4U,  hIDAT, hIHDR,        1
-   /* Supported chunks from PNG extensions 1.5.0, NYI so limit */
-#  define CDoFFs       9U,    9U,  hIDAT, hIHDR,        0
-#  define CDpCAL  NoCheck,   14U,  hIDAT, hIHDR,        0
-      /* Allocates 'length+1'; checked in the handler */
-#  define CDsCAL    Limit,    4U,  hIDAT, hIHDR,        0
-      /* Allocates 'length+1'; checked in the handler */
-
-#  define PNG_CHUNK(cHNK, index) { png_handle_ ## cHNK, CD ## cHNK },
-   PNG_KNOWN_CHUNKS
-#  undef PNG_CHUNK
-};
-
-
-static png_index
-png_chunk_index_from_name(png_uint_32 chunk_name)
-{
-   /* For chunk png_cHNK return PNG_INDEX_cHNK.  Return PNG_INDEX_unknown if
-    * chunk_name is not known.  Notice that in a particular build "known" does
-    * not necessarily mean "supported", although the inverse applies.
-    */
-   switch (chunk_name)
-   {
-#     define PNG_CHUNK(cHNK, index)\
-         case png_ ## cHNK: return PNG_INDEX_ ## cHNK; /* == index */
-
-      PNG_KNOWN_CHUNKS
-
-#     undef PNG_CHUNK
-
-      default: return PNG_INDEX_unknown;
-   }
-}
 
-png_handle_result_code /*PRIVATE*/
-png_handle_chunk(png_structrp png_ptr, png_inforp info_ptr, png_uint_32 length)
+void /* PRIVATE */
+png_check_chunk_name(png_const_structrp png_ptr, png_uint_32 chunk_name)
 {
-   /* CSE: these things don't change, these autos are just to save typing and
-    * make the code more clear.
-    */
-   const png_uint_32 chunk_name = png_ptr->chunk_name;
-   const png_index chunk_index = png_chunk_index_from_name(chunk_name);
-
-   png_handle_result_code handled = handled_error;
-   png_const_charp errmsg = NULL;
-
-   /* Is this a known chunk?  If not there are no checks performed here;
-    * png_handle_unknown does the correct checks.  This means that the values
-    * for known but unsupported chunks in the above table are not used here
-    * however the chunks_seen fields in png_struct are still set.
-    */
-   if (chunk_index == PNG_INDEX_unknown ||
-       read_chunks[chunk_index].handler == NULL)
-   {
-      handled = png_handle_unknown(
-            png_ptr, info_ptr, length, PNG_HANDLE_CHUNK_AS_DEFAULT);
-   }
-
-   /* First check the position.   The first check is historical; the stream must
-    * start with IHDR and anything else causes libpng to give up immediately.
-    */
-   else if (chunk_index != PNG_INDEX_IHDR &&
-            (png_ptr->mode & PNG_HAVE_IHDR) == 0)
-      png_chunk_error(png_ptr, "missing IHDR"); /* NORETURN */
-
-   /* Before all the pos_before chunks, after all the pos_after chunks. */
-   else if (((png_ptr->mode & read_chunks[chunk_index].pos_before) != 0) ||
-            ((png_ptr->mode & read_chunks[chunk_index].pos_after) !=
-             read_chunks[chunk_index].pos_after))
-   {
-      errmsg = "out of place";
-   }
+   int i;
+   png_uint_32 cn=chunk_name;
 
-   /* Now check for duplicates: duplicated critical chunks also produce a
-    * full error.
-    */
-   else if (read_chunks[chunk_index].multiple == 0 &&
-            png_file_has_chunk(png_ptr, chunk_index))
-   {
-      errmsg = "duplicate";
-   }
+   png_debug(1, "in png_check_chunk_name");
 
-   else if (length < read_chunks[chunk_index].min_length)
-      errmsg = "too short";
-   else
+   for (i=1; i<=4; ++i)
    {
-      /* NOTE: apart from IHDR the critical chunks (PLTE, IDAT and IEND) are set
-       * up above not to do any length checks.
-       *
-       * The png_chunk_max check ensures that the variable length chunks are
-       * always checked at this point for being within the system allocation
-       * limits.
-       */
-      unsigned max_length = read_chunks[chunk_index].max_length;
-
-      switch (max_length)
-      {
-         case Limit:
-            /* png_read_chunk_header has already png_error'ed chunks with a
-             * length exceeding the 31-bit PNG limit, so just check the memory
-             * limit:
-             */
-            if (length <= png_chunk_max(png_ptr))
-               goto MeetsLimit;
+      int c = cn & 0xff;
 
-            errmsg = "length exceeds libpng limit";
-            break;
-
-         default:
-            if (length <= max_length)
-               goto MeetsLimit;
-
-            errmsg = "too long";
-            break;
+      if (c < 65 || c > 122 || (c > 90 && c < 97))
+         png_chunk_error(png_ptr, "invalid chunk type");
 
-         case NoCheck:
-         MeetsLimit:
-            handled = read_chunks[chunk_index].handler(
-                  png_ptr, info_ptr, length);
-            break;
-      }
+      cn >>= 8;
    }
+}
 
-   /* If there was an error or the chunk was simply skipped it is not counted as
-    * 'seen'.
-    */
-   if (errmsg != NULL)
+void /* PRIVATE */
+png_check_chunk_length(png_const_structrp png_ptr, png_uint_32 length)
+{
+   png_alloc_size_t limit = PNG_UINT_31_MAX;
+
+# ifdef PNG_SET_USER_LIMITS_SUPPORTED
+   if (png_ptr->user_chunk_malloc_max > 0 &&
+       png_ptr->user_chunk_malloc_max < limit)
+      limit = png_ptr->user_chunk_malloc_max;
+# elif PNG_USER_CHUNK_MALLOC_MAX > 0
+   if (PNG_USER_CHUNK_MALLOC_MAX < limit)
+      limit = PNG_USER_CHUNK_MALLOC_MAX;
+# endif
+   if (png_ptr->chunk_name == png_IDAT)
    {
-      if (PNG_CHUNK_CRITICAL(chunk_name)) /* stop immediately */
-         png_chunk_error(png_ptr, errmsg);
-      else /* ancillary chunk */
-      {
-         /* The chunk data is skipped: */
-         png_crc_finish(png_ptr, length);
-         png_chunk_benign_error(png_ptr, errmsg);
-      }
+      png_alloc_size_t idat_limit = PNG_UINT_31_MAX;
+      size_t row_factor =
+         (size_t)png_ptr->width
+         * (size_t)png_ptr->channels
+         * (png_ptr->bit_depth > 8? 2: 1)
+         + 1
+         + (png_ptr->interlaced? 6: 0);
+      if (png_ptr->height > PNG_UINT_32_MAX/row_factor)
+         idat_limit = PNG_UINT_31_MAX;
+      else
+         idat_limit = png_ptr->height * row_factor;
+      row_factor = row_factor > 32566? 32566 : row_factor;
+      idat_limit += 6 + 5*(idat_limit/row_factor+1); /* zlib+deflate overhead */
+      idat_limit=idat_limit < PNG_UINT_31_MAX? idat_limit : PNG_UINT_31_MAX;
+      limit = limit < idat_limit? idat_limit : limit;
    }
 
-   else if (handled >= handled_saved)
+   if (length > limit)
    {
-      if (chunk_index != PNG_INDEX_unknown)
-         png_file_add_chunk(png_ptr, chunk_index);
+      png_debug2(0," length = %lu, limit = %lu",
+         (unsigned long)length,(unsigned long)limit);
+      png_benign_error(png_ptr, "chunk data is too large");
    }
-
-   return handled;
 }
 
 /* Combines the row recently read in with the existing pixels in the row.  This
@@ -3723,6 +3684,10 @@ void /* PRIVATE */
 png_do_read_interlace(png_row_infop row_info, png_bytep row, int pass,
     png_uint_32 transformations /* Because these may affect the byte layout */)
 {
+   /* Arrays to facilitate easy interlacing - use pass (0 - 6) as index */
+   /* Offset to next interlace block */
+   static const unsigned int png_pass_inc[7] = {8, 8, 4, 4, 2, 2, 1};
+
    png_debug(1, "in png_do_read_interlace");
    if (row != NULL && row_info != NULL)
    {
@@ -4215,9 +4180,6 @@ png_read_IDAT_data(png_structrp png_ptr, png_bytep output,
 
          avail_in = png_ptr->IDAT_read_size;
 
-         if (avail_in > png_chunk_max(png_ptr))
-            avail_in = (uInt)/*SAFE*/png_chunk_max(png_ptr);
-
          if (avail_in > png_ptr->idat_size)
             avail_in = (uInt)png_ptr->idat_size;
 
@@ -4225,13 +4187,8 @@ png_read_IDAT_data(png_structrp png_ptr, png_bytep output,
           * to minimize memory usage by causing lots of re-allocs, but
           * realistically doing IDAT_read_size re-allocs is not likely to be a
           * big problem.
-          *
-          * An error here corresponds to the system being out of memory.
           */
-         buffer = png_read_buffer(png_ptr, avail_in);
-
-         if (buffer == NULL)
-            png_chunk_error(png_ptr, "out of memory");
+         buffer = png_read_buffer(png_ptr, avail_in, 0/*error*/);
 
          png_crc_read(png_ptr, buffer, avail_in);
          png_ptr->idat_size -= avail_in;
@@ -4368,6 +4325,20 @@ png_read_finish_IDAT(png_structrp png_ptr)
 void /* PRIVATE */
 png_read_finish_row(png_structrp png_ptr)
 {
+   /* Arrays to facilitate easy interlacing - use pass (0 - 6) as index */
+
+   /* Start of interlace block */
+   static const png_byte png_pass_start[7] = {0, 4, 0, 2, 0, 1, 0};
+
+   /* Offset to next interlace block */
+   static const png_byte png_pass_inc[7] = {8, 8, 4, 4, 2, 2, 1};
+
+   /* Start of interlace block in the y direction */
+   static const png_byte png_pass_ystart[7] = {0, 0, 4, 0, 2, 0, 1};
+
+   /* Offset to next interlace block in the y direction */
+   static const png_byte png_pass_yinc[7] = {8, 8, 8, 4, 4, 2, 2};
+
    png_debug(1, "in png_read_finish_row");
    png_ptr->row_number++;
    if (png_ptr->row_number < png_ptr->num_rows)
@@ -4419,6 +4390,20 @@ png_read_finish_row(png_structrp png_ptr)
 void /* PRIVATE */
 png_read_start_row(png_structrp png_ptr)
 {
+   /* Arrays to facilitate easy interlacing - use pass (0 - 6) as index */
+
+   /* Start of interlace block */
+   static const png_byte png_pass_start[7] = {0, 4, 0, 2, 0, 1, 0};
+
+   /* Offset to next interlace block */
+   static const png_byte png_pass_inc[7] = {8, 8, 4, 4, 2, 2, 1};
+
+   /* Start of interlace block in the y direction */
+   static const png_byte png_pass_ystart[7] = {0, 0, 4, 0, 2, 0, 1};
+
+   /* Offset to next interlace block in the y direction */
+   static const png_byte png_pass_yinc[7] = {8, 8, 8, 4, 4, 2, 2};
+
    unsigned int max_pixel_depth;
    size_t row_bytes;
 
