diff --git a/libs/png/pngwio.c b/libs/png/pngwio.c
index 96a3187ff0139172f7274e2db630f4f47314fce6..10e919dd03819f62ba53a85acfcb4e4be176d9c8 100644
--- a/libs/png/pngwio.c
+++ b/libs/png/pngwio.c
@@ -1,6 +1,7 @@
+
 /* pngwio.c - functions for data output
  *
- * Copyright (c) 2018-2025 Cosmin Truta
+ * Copyright (c) 2018 Cosmin Truta
  * Copyright (c) 1998-2002,2004,2006-2014,2016,2018 Glenn Randers-Pehrson
  * Copyright (c) 1996-1997 Andreas Dilger
  * Copyright (c) 1995-1996 Guy Eric Schalnat, Group 42, Inc.
@@ -54,7 +55,7 @@ png_default_write_data(png_structp png_ptr, png_bytep data, size_t length)
    if (png_ptr == NULL)
       return;
 
-   check = fwrite(data, 1, length, (FILE *)png_ptr->io_ptr);
+   check = fwrite(data, 1, length, (png_FILE_p)(png_ptr->io_ptr));
 
    if (check != length)
       png_error(png_ptr, "Write Error");
@@ -77,12 +78,12 @@ png_flush(png_structrp png_ptr)
 void PNGCBAPI
 png_default_flush(png_structp png_ptr)
 {
-   FILE *io_ptr;
+   png_FILE_p io_ptr;
 
    if (png_ptr == NULL)
       return;
 
-   io_ptr = png_voidcast(FILE *, png_ptr->io_ptr);
+   io_ptr = png_voidcast(png_FILE_p, (png_ptr->io_ptr));
    fflush(io_ptr);
 }
 #  endif
