diff --git a/libs/strmbase/filter.c b/libs/strmbase/filter.c
index 62cfaf9e6c0a897307a2ee03c05e67a81764fee2..91e459c18ada8af6a1047ef5f721db7b6542ad7f 100644
--- a/libs/strmbase/filter.c
+++ b/libs/strmbase/filter.c
@@ -18,9 +18,7 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#define COBJMACROS
-#include "wine/strmbase.h"
-#include "wine/debug.h"
+#include "strmbase_private.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(quartz);
 
@@ -43,7 +41,7 @@ static HRESULT enum_pins_create(struct strmbase_filter *filter, IEnumPins **out)
     if (!out)
         return E_POINTER;
 
-    if (!(object = calloc(1, sizeof(*object))))
+    if (!(object = heap_alloc_zero(sizeof(*object))))
     {
         *out = NULL;
         return E_OUTOFMEMORY;
@@ -102,7 +100,7 @@ static ULONG WINAPI enum_pins_Release(IEnumPins *iface)
     if (!refcount)
     {
         IBaseFilter_Release(&enum_pins->filter->IBaseFilter_iface);
-        free(enum_pins);
+        heap_free(enum_pins);
     }
     return refcount;
 }
@@ -228,6 +226,8 @@ static HRESULT WINAPI filter_inner_QueryInterface(IUnknown *iface, REFIID iid, v
     {
         *out = &filter->IBaseFilter_iface;
     }
+    else if (IsEqualIID(iid, &IID_IPropertyBag))
+        *out = &filter->IPropertyBag_iface;
     else
     {
         WARN("%s not implemented, returning E_NOINTERFACE.\n", debugstr_guid(iid));
@@ -511,6 +511,51 @@ static const IBaseFilterVtbl filter_vtbl =
     filter_QueryVendorInfo,
 };
 
+static inline struct strmbase_filter *impl_from_IPropertyBag(IPropertyBag *iface)
+{
+    return CONTAINING_RECORD(iface, struct strmbase_filter, IPropertyBag_iface);
+}
+
+static HRESULT WINAPI property_bag_QueryInterface(IPropertyBag *iface, REFIID iid, void **out)
+{
+    struct strmbase_filter *filter = impl_from_IPropertyBag(iface);
+    return IUnknown_QueryInterface(filter->outer_unk, iid, out);
+}
+
+static ULONG WINAPI property_bag_AddRef(IPropertyBag *iface)
+{
+    struct strmbase_filter *filter = impl_from_IPropertyBag(iface);
+    return IUnknown_AddRef(filter->outer_unk);
+}
+
+static ULONG WINAPI property_bag_Release(IPropertyBag *iface)
+{
+    struct strmbase_filter *filter = impl_from_IPropertyBag(iface);
+    return IUnknown_Release(filter->outer_unk);
+}
+
+static HRESULT WINAPI property_bag_Read(IPropertyBag *iface, const WCHAR *prop_name, VARIANT *value,
+        IErrorLog *error_log)
+{
+    FIXME("iface %p, prop_name %s, value %p, error_log %p stub!\n", iface, debugstr_w(prop_name), value, error_log);
+    return E_NOTIMPL;
+}
+
+static HRESULT WINAPI property_bag_Write(IPropertyBag *iface, const WCHAR *prop_name, VARIANT *value)
+{
+    FIXME("iface %p, prop_name %s, value %p stub!\n", iface, debugstr_w(prop_name), value);
+    return S_OK;
+}
+
+static const IPropertyBagVtbl property_bag_vtbl =
+{
+    property_bag_QueryInterface,
+    property_bag_AddRef,
+    property_bag_Release,
+    property_bag_Read,
+    property_bag_Write,
+};
+
 VOID WINAPI BaseFilterImpl_IncrementPinVersion(struct strmbase_filter *filter)
 {
     InterlockedIncrement(&filter->pin_version);
@@ -522,6 +567,7 @@ void strmbase_filter_init(struct strmbase_filter *filter, IUnknown *outer,
     memset(filter, 0, sizeof(*filter));
 
     filter->IBaseFilter_iface.lpVtbl = &filter_vtbl;
+    filter->IPropertyBag_iface.lpVtbl = &property_bag_vtbl;
     filter->IUnknown_inner.lpVtbl = &filter_inner_vtbl;
     filter->outer_unk = outer ? outer : &filter->IUnknown_inner;
     filter->refcount = 1;
