diff --git a/libs/strmbase/mediatype.c b/libs/strmbase/mediatype.c
index e551aeb06cf1f8ffdf5cbd68eb7004929d545d75..af813e2ffee0fa24e6b07c2daf13ef0536126515 100644
--- a/libs/strmbase/mediatype.c
+++ b/libs/strmbase/mediatype.c
@@ -19,11 +19,9 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#define COBJMACROS
-#include "wine/strmbase.h"
+#include "strmbase_private.h"
 #include "dvdmedia.h"
 #include "dxva.h"
-#include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(quartz);
 
