diff --git a/libs/strmbase/pospass.c b/libs/strmbase/pospass.c
index 3a358475a409fbef018e8cfc43b8fa8e9c80b3ce..2b6321432bd6fe04caddf28b8df46d2c69e06aa1 100644
--- a/libs/strmbase/pospass.c
+++ b/libs/strmbase/pospass.c
@@ -20,9 +20,7 @@
  */
 /* FIXME: critical sections */
 
-#define COBJMACROS
-#include "wine/strmbase.h"
-#include "wine/debug.h"
+#include "strmbase_private.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(quartz);
 
