diff --git a/libs/strmbase/renderer.c b/libs/strmbase/renderer.c
index 7305599ff78319b779b6c049cdf4fde4d37481a4..dc421f499edd335e3f416309e3e01a1e28382344 100644
--- a/libs/strmbase/renderer.c
+++ b/libs/strmbase/renderer.c
@@ -18,9 +18,7 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#define COBJMACROS
-#include "wine/strmbase.h"
-#include "wine/debug.h"
+#include "strmbase_private.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(quartz);
 
