diff --git a/libs/strmbase/seeking.c b/libs/strmbase/seeking.c
index cc602f0a44d0742ed6ac93e14c75ea396b63a4e2..62e6d454a57e8aa748d181b649124285da746e1a 100644
--- a/libs/strmbase/seeking.c
+++ b/libs/strmbase/seeking.c
@@ -19,10 +19,7 @@
  * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
  */
 
-#include <assert.h>
-#define COBJMACROS
-#include "wine/strmbase.h"
-#include "wine/debug.h"
+#include "strmbase_private.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(quartz);
 
