diff --git a/libs/tiff/libtiff/tif_aux.c b/libs/tiff/libtiff/tif_aux.c
index 31b240070c65ea919438adac6c248a7986aa3188..e9606a42e0fe24ebcc5203ae50f27c43848cc2d5 100644
--- a/libs/tiff/libtiff/tif_aux.c
+++ b/libs/tiff/libtiff/tif_aux.c
@@ -131,10 +131,7 @@ static int TIFFDefaultTransferFunction(TIFF *tif, TIFFDirectory *td)
     tmsize_t i, n, nbytes;
 
     tf[0] = tf[1] = tf[2] = 0;
-    // Do not try to generate a default TransferFunction beyond 24 bits.
-    // This otherwise leads to insane amounts, resulting in denial of service
-    // See https://github.com/OSGeo/gdal/issues/10875
-    if (td->td_bitspersample > 24)
+    if (td->td_bitspersample >= sizeof(tmsize_t) * 8 - 2)
         return 0;
 
     n = ((tmsize_t)1) << td->td_bitspersample;
