diff --git a/libs/tiff/libtiff/tif_close.c b/libs/tiff/libtiff/tif_close.c
index d498010740c570aa07637e4c2ae2d046f2f1a205..d6bb7f1d7cc6216b18c9c27cc13c2980ca148ba7 100644
--- a/libs/tiff/libtiff/tif_close.c
+++ b/libs/tiff/libtiff/tif_close.c
@@ -48,6 +48,7 @@ void TIFFCleanup(TIFF *tif)
      */
     if (tif->tif_mode != O_RDONLY)
         TIFFFlush(tif);
+    (*tif->tif_cleanup)(tif);
     TIFFFreeDirectory(tif);
 
     _TIFFCleanupIFDOffsetAndNumberMaps(tif);
