diff --git a/libs/tiff/libtiff/tif_dirread.c b/libs/tiff/libtiff/tif_dirread.c
index 7f0aeb1544c30a257765477aead89fac004283fb..3d94ea7d4f37cd372d66092a4ff9e6fee1849953 100644
--- a/libs/tiff/libtiff/tif_dirread.c
+++ b/libs/tiff/libtiff/tif_dirread.c
@@ -47,9 +47,6 @@
 #define TIFFCvtIEEEFloatToNative(tif, n, fp)
 #define TIFFCvtIEEEDoubleToNative(tif, n, dp)
 #else
-/* If your machine does not support IEEE floating point then you will need to
- * add support to tif_machdep.c to convert between the native format and
- * IEEE format. */
 extern void TIFFCvtIEEEFloatToNative(TIFF *, uint32_t, float *);
 extern void TIFFCvtIEEEDoubleToNative(TIFF *, uint32_t, double *);
 #endif
@@ -2784,7 +2781,7 @@ TIFFReadDirEntryFloatArray(TIFF *tif, TIFFDirEntry *direntry, float **value)
         case TIFF_FLOAT:
             if (tif->tif_flags & TIFF_SWAB)
                 TIFFSwabArrayOfLong((uint32_t *)origdata, count);
-            TIFFCvtIEEEFloatToNative(tif, count, (float *)origdata);
+            TIFFCvtIEEEDoubleToNative(tif, count, (float *)origdata);
             *value = (float *)origdata;
             return (TIFFReadDirEntryErrOk);
     }
@@ -4313,6 +4310,7 @@ int TIFFReadDirectory(TIFF *tif)
     tif->tif_flags &= ~TIFF_CHOPPEDUPARRAYS;
 
     /* free any old stuff and reinit */
+    (*tif->tif_cleanup)(tif); /* cleanup any previous compression state */
     TIFFFreeDirectory(tif);
     TIFFDefaultDirectory(tif);
 
@@ -4405,13 +4403,10 @@ int TIFFReadDirectory(TIFF *tif)
             TIFFReadDirectoryFindFieldInfo(tif, dp->tdir_tag, &fii);
             if (fii == FAILED_FII)
             {
-                if (tif->tif_warn_about_unknown_tags)
-                {
-                    TIFFWarningExtR(tif, module,
-                                    "Unknown field with tag %" PRIu16
-                                    " (0x%" PRIx16 ") encountered",
-                                    dp->tdir_tag, dp->tdir_tag);
-                }
+                TIFFWarningExtR(tif, module,
+                                "Unknown field with tag %" PRIu16 " (0x%" PRIx16
+                                ") encountered",
+                                dp->tdir_tag, dp->tdir_tag);
                 /* the following knowingly leaks the
                    anonymous field structure */
                 const TIFFField *fld = _TIFFCreateAnonField(
@@ -5286,7 +5281,6 @@ int TIFFReadCustomDirectory(TIFF *tif, toff_t diroff,
     const TIFFField *fip;
     uint32_t fii;
 
-    assert(infoarray != NULL);
     dircount = TIFFFetchDirectory(tif, diroff, &dir, NULL);
     if (!dircount)
     {
@@ -5319,6 +5313,7 @@ int TIFFReadCustomDirectory(TIFF *tif, toff_t diroff,
     }
 
     /* Free any old stuff and reinit. */
+    (*tif->tif_cleanup)(tif); /* cleanup any previous compression state */
     TIFFFreeDirectory(tif);
     /* Even if custom directories do not need the default settings of a standard
      * IFD, the pointer to the TIFFSetField() and TIFFGetField() (i.e.
@@ -5349,25 +5344,18 @@ int TIFFReadCustomDirectory(TIFF *tif, toff_t diroff,
         TIFFReadDirectoryFindFieldInfo(tif, dp->tdir_tag, &fii);
         if (fii == FAILED_FII)
         {
-            if (tif->tif_warn_about_unknown_tags)
-            {
-                TIFFWarningExtR(tif, module,
-                                "Unknown field with tag %" PRIu16 " (0x%" PRIx16
-                                ") encountered",
-                                dp->tdir_tag, dp->tdir_tag);
-            }
+            TIFFWarningExtR(tif, module,
+                            "Unknown field with tag %" PRIu16 " (0x%" PRIx16
+                            ") encountered",
+                            dp->tdir_tag, dp->tdir_tag);
             const TIFFField *fld = _TIFFCreateAnonField(
                 tif, dp->tdir_tag, (TIFFDataType)dp->tdir_type);
             if (fld == NULL || !_TIFFMergeFields(tif, fld, 1))
             {
-                if (tif->tif_warn_about_unknown_tags)
-                {
-                    TIFFWarningExtR(
-                        tif, module,
-                        "Registering anonymous field with tag %" PRIu16
-                        " (0x%" PRIx16 ") failed",
-                        dp->tdir_tag, dp->tdir_tag);
-                }
+                TIFFWarningExtR(tif, module,
+                                "Registering anonymous field with tag %" PRIu16
+                                " (0x%" PRIx16 ") failed",
+                                dp->tdir_tag, dp->tdir_tag);
                 dp->tdir_ignore = TRUE;
             }
             else
@@ -6285,19 +6273,19 @@ static int TIFFFetchNormalTag(TIFF *tif, TIFFDirEntry *dp, int recover)
     }
     fip = tif->tif_fields[fii];
     assert(fip != NULL); /* should not happen */
-    assert(fip->set_get_field_type !=
+    assert(fip->set_field_type !=
            TIFF_SETGET_OTHER); /* if so, we shouldn't arrive here but deal with
                                   this in specialized code */
-    assert(fip->set_get_field_type !=
+    assert(fip->set_field_type !=
            TIFF_SETGET_INT); /* if so, we shouldn't arrive here as this is only
                                 the case for pseudo-tags */
     err = TIFFReadDirEntryErrOk;
-    switch (fip->set_get_field_type)
+    switch (fip->set_field_type)
     {
         case TIFF_SETGET_UNDEFINED:
             TIFFErrorExtR(
                 tif, "TIFFFetchNormalTag",
-                "Defined set_get_field_type of custom tag %u (%s) is "
+                "Defined set_field_type of custom tag %u (%s) is "
                 "TIFF_SETGET_UNDEFINED and thus tag is not read from file",
                 fip->field_tag, fip->field_name);
             break;
