diff --git a/libs/tiff/libtiff/tif_fax3.c b/libs/tiff/libtiff/tif_fax3.c
index 34e1c7882f772610b35fd4dcf5ef2d69b129a405..01a784730bdcc4b68e08e9aa49581c8aa0485f29 100644
--- a/libs/tiff/libtiff/tif_fax3.c
+++ b/libs/tiff/libtiff/tif_fax3.c
@@ -87,10 +87,6 @@ typedef struct
     int EOLcnt;                  /* count of EOL codes recognized */
     int eofReachedCount;         /* number of times decode has been called with
                                     EOF already reached */
-    int eolReachedCount;         /* number of times decode has been called with
-                                    EOL already reached */
-    int unexpectedReachedCount;  /* number of times decode has been called with
-                                    "unexpedted" already reached */
     TIFFFaxFillFunc fill;        /* fill routine */
     uint32_t *runs;              /* b&w runs for current/previous row */
     uint32_t nruns;              /* size of the refruns / curruns arrays */
@@ -178,8 +174,6 @@ static int Fax3PreDecode(TIFF *tif, uint16_t s)
     sp->data = 0;
     sp->EOLcnt = 0; /* force initial scan for EOL */
     sp->eofReachedCount = 0;
-    sp->eolReachedCount = 0;
-    sp->unexpectedReachedCount = 0;
     /*
      * Decoder assumes lsb-to-msb bit order.  Note that we select
      * this here rather than in Fax3SetupState so that viewers can
@@ -215,12 +209,7 @@ static void Fax3Unexpected(const char *module, TIFF *tif, uint32_t line,
                   line, isTiled(tif) ? "tile" : "strip",
                   (isTiled(tif) ? tif->tif_curtile : tif->tif_curstrip), a0);
 }
-#define unexpected(table, a0)                                                  \
-    do                                                                         \
-    {                                                                          \
-        Fax3Unexpected(module, tif, sp->line, a0);                             \
-        ++sp->unexpectedReachedCount;                                          \
-    } while (0)
+#define unexpected(table, a0) Fax3Unexpected(module, tif, sp->line, a0)
 
 static void Fax3Extension(const char *module, TIFF *tif, uint32_t line,
                           uint32_t a0)
@@ -244,12 +233,7 @@ static void Fax3BadLength(const char *module, TIFF *tif, uint32_t line,
                     (isTiled(tif) ? tif->tif_curtile : tif->tif_curstrip), a0,
                     lastx);
 }
-#define badlength(a0, lastx)                                                   \
-    do                                                                         \
-    {                                                                          \
-        Fax3BadLength(module, tif, sp->line, a0, lastx);                       \
-        ++sp->eolReachedCount;                                                 \
-    } while (0)
+#define badlength(a0, lastx) Fax3BadLength(module, tif, sp->line, a0, lastx)
 
 static void Fax3PrematureEOF(const char *module, TIFF *tif, uint32_t line,
                              uint32_t a0)
@@ -267,55 +251,8 @@ static void Fax3PrematureEOF(const char *module, TIFF *tif, uint32_t line,
         ++sp->eofReachedCount;                                                 \
     } while (0)
 
-static void Fax3TryG3WithoutEOL(const char *module, TIFF *tif, uint32_t line,
-                                uint32_t a0)
-{
-    TIFFWarningExtR(
-        tif, module,
-        "Try to decode (read) fax Group 3 data without EOL at line %" PRIu32
-        " of %s %" PRIu32 " (x %" PRIu32 "). Please check result",
-        line, isTiled(tif) ? "tile" : "strip",
-        (isTiled(tif) ? tif->tif_curtile : tif->tif_curstrip), a0);
-}
-#define tryG3WithoutEOL(a0)                                                    \
-    do                                                                         \
-    {                                                                          \
-        Fax3TryG3WithoutEOL(module, tif, sp->line, a0);                        \
-    } while (0)
-
 #define Nop
 
-static int CheckReachedCounters(TIFF *tif, const char *module,
-                                Fax3CodecState *sp)
-{
-    if (sp->eofReachedCount >= EOF_REACHED_COUNT_THRESHOLD)
-    {
-        TIFFErrorExtR(tif, module,
-                      "End of file (EOF) has already been reached %d times "
-                      "within that %s.",
-                      sp->eofReachedCount, isTiled(tif) ? "tile" : "strip");
-        return (-1);
-    }
-    if (sp->eolReachedCount >= EOF_REACHED_COUNT_THRESHOLD)
-    {
-        TIFFErrorExtR(tif, module,
-                      "Bad line length (EOL) has already been reached %d times "
-                      "within that %s",
-                      sp->eolReachedCount, isTiled(tif) ? "tile" : "strip");
-        return (-1);
-    }
-    if (sp->unexpectedReachedCount >= EOF_REACHED_COUNT_THRESHOLD)
-    {
-        TIFFErrorExtR(tif, module,
-                      "Bad code word (unexpected) has already been reached %d "
-                      "times within that %s",
-                      sp->unexpectedReachedCount,
-                      isTiled(tif) ? "tile" : "strip");
-        return (-1);
-    }
-    return (0);
-}
-
 /**
  * Decode the requested amount of G3 1D-encoded data.
  * @param buf destination buffer
@@ -332,9 +269,14 @@ static int Fax3Decode1D(TIFF *tif, uint8_t *buf, tmsize_t occ, uint16_t s)
         TIFFErrorExtR(tif, module, "Fractional scanlines cannot be read");
         return (-1);
     }
-    if (CheckReachedCounters(tif, module, sp))
+    if (sp->eofReachedCount >= EOF_REACHED_COUNT_THRESHOLD)
+    {
+        TIFFErrorExtR(
+            tif, module,
+            "End of file has already been reached %d times within that strip",
+            sp->eofReachedCount);
         return (-1);
-RETRY_WITHOUT_EOL_1D:
+    }
     CACHE_STATE(tif, sp);
     thisrun = sp->curruns;
     while (occ > 0)
@@ -347,7 +289,7 @@ RETRY_WITHOUT_EOL_1D:
         printf("-------------------- %" PRIu32 "\n", tif->tif_row);
         fflush(stdout);
 #endif
-        SYNC_EOL(EOF1D, RETRY_WITHOUT_EOL_1D);
+        SYNC_EOL(EOF1D);
         EXPAND1D(EOF1Da);
         (*sp->fill)(buf, thisrun, pa, lastx);
         buf += sp->b.rowbytes;
@@ -385,9 +327,14 @@ static int Fax3Decode2D(TIFF *tif, uint8_t *buf, tmsize_t occ, uint16_t s)
         TIFFErrorExtR(tif, module, "Fractional scanlines cannot be read");
         return (-1);
     }
-    if (CheckReachedCounters(tif, module, sp))
+    if (sp->eofReachedCount >= EOF_REACHED_COUNT_THRESHOLD)
+    {
+        TIFFErrorExtR(
+            tif, module,
+            "End of file has already been reached %d times within that strip",
+            sp->eofReachedCount);
         return (-1);
-RETRY_WITHOUT_EOL_2D:
+    }
     CACHE_STATE(tif, sp);
     while (occ > 0)
     {
@@ -398,7 +345,7 @@ RETRY_WITHOUT_EOL_2D:
         printf("\nBitAcc=%08" PRIX32 ", BitsAvail = %d EOLcnt = %d", BitAcc,
                BitsAvail, EOLcnt);
 #endif
-        SYNC_EOL(EOF2D, RETRY_WITHOUT_EOL_2D);
+        SYNC_EOL(EOF2D);
         NeedBits8(1, EOF2D);
         is1D = GetBits(1); /* 1D/2D-encoding tag bit */
         ClrBits(1);
@@ -588,15 +535,6 @@ static int Fax3SetupState(TIFF *tif)
                       "Bits/sample must be 1 for Group 3/4 encoding/decoding");
         return (0);
     }
-    if (td->td_samplesperpixel != 1 &&
-        td->td_planarconfig != PLANARCONFIG_SEPARATE)
-    {
-        TIFFErrorExtR(
-            tif, module,
-            "Samples/pixel shall be 1 for Group 3/4 encoding/decoding, "
-            "or PlanarConfiguration must be set to Separate.");
-        return 0;
-    }
     /*
      * Calculate the scanline/tile widths.
      */
@@ -1337,23 +1275,24 @@ static void Fax3Cleanup(TIFF *tif)
 #define FIELD_OPTIONS (FIELD_CODEC + 7)
 
 static const TIFFField faxFields[] = {
-    {TIFFTAG_FAXMODE, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT, FIELD_PSEUDO, FALSE,
-     FALSE, "FaxMode", NULL},
-    {TIFFTAG_FAXFILLFUNC, 0, 0, TIFF_ANY, 0, TIFF_SETGET_OTHER, FIELD_PSEUDO,
-     FALSE, FALSE, "FaxFillFunc", NULL},
+    {TIFFTAG_FAXMODE, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT, TIFF_SETGET_UNDEFINED,
+     FIELD_PSEUDO, FALSE, FALSE, "FaxMode", NULL},
+    {TIFFTAG_FAXFILLFUNC, 0, 0, TIFF_ANY, 0, TIFF_SETGET_OTHER,
+     TIFF_SETGET_UNDEFINED, FIELD_PSEUDO, FALSE, FALSE, "FaxFillFunc", NULL},
     {TIFFTAG_BADFAXLINES, 1, 1, TIFF_LONG, 0, TIFF_SETGET_UINT32,
-     FIELD_BADFAXLINES, TRUE, FALSE, "BadFaxLines", NULL},
+     TIFF_SETGET_UINT32, FIELD_BADFAXLINES, TRUE, FALSE, "BadFaxLines", NULL},
     {TIFFTAG_CLEANFAXDATA, 1, 1, TIFF_SHORT, 0, TIFF_SETGET_UINT16,
-     FIELD_CLEANFAXDATA, TRUE, FALSE, "CleanFaxData", NULL},
+     TIFF_SETGET_UINT16, FIELD_CLEANFAXDATA, TRUE, FALSE, "CleanFaxData", NULL},
     {TIFFTAG_CONSECUTIVEBADFAXLINES, 1, 1, TIFF_LONG, 0, TIFF_SETGET_UINT32,
-     FIELD_BADFAXRUN, TRUE, FALSE, "ConsecutiveBadFaxLines", NULL}};
+     TIFF_SETGET_UINT32, FIELD_BADFAXRUN, TRUE, FALSE, "ConsecutiveBadFaxLines",
+     NULL}};
 static const TIFFField fax3Fields[] = {
     {TIFFTAG_GROUP3OPTIONS, 1, 1, TIFF_LONG, 0, TIFF_SETGET_UINT32,
-     FIELD_OPTIONS, FALSE, FALSE, "Group3Options", NULL},
+     TIFF_SETGET_UINT32, FIELD_OPTIONS, FALSE, FALSE, "Group3Options", NULL},
 };
 static const TIFFField fax4Fields[] = {
     {TIFFTAG_GROUP4OPTIONS, 1, 1, TIFF_LONG, 0, TIFF_SETGET_UINT32,
-     FIELD_OPTIONS, FALSE, FALSE, "Group4Options", NULL},
+     TIFF_SETGET_UINT32, FIELD_OPTIONS, FALSE, FALSE, "Group4Options", NULL},
 };
 
 static int Fax3VSetField(TIFF *tif, uint32_t tag, va_list ap)
@@ -1616,8 +1555,14 @@ static int Fax4Decode(TIFF *tif, uint8_t *buf, tmsize_t occ, uint16_t s)
         TIFFErrorExtR(tif, module, "Fractional scanlines cannot be read");
         return (-1);
     }
-    if (CheckReachedCounters(tif, module, sp))
+    if (sp->eofReachedCount >= EOF_REACHED_COUNT_THRESHOLD)
+    {
+        TIFFErrorExtR(
+            tif, module,
+            "End of file has already been reached %d times within that strip",
+            sp->eofReachedCount);
         return (-1);
+    }
     CACHE_STATE(tif, sp);
     int start = sp->line;
     while (occ > 0)
@@ -1762,8 +1707,6 @@ static int Fax3DecodeRLE(TIFF *tif, uint8_t *buf, tmsize_t occ, uint16_t s)
         TIFFErrorExtR(tif, module, "Fractional scanlines cannot be read");
         return (-1);
     }
-    if (CheckReachedCounters(tif, module, sp))
-        return (-1);
     CACHE_STATE(tif, sp);
     thisrun = sp->curruns;
     while (occ > 0)
