diff --git a/libs/tiff/libtiff/tif_fax3.h b/libs/tiff/libtiff/tif_fax3.h
index c3100ce2c6c6c196ea6e9919ebff443efc84b72c..e095009bba4a99e01d8ac80678606b68278c267a 100644
--- a/libs/tiff/libtiff/tif_fax3.h
+++ b/libs/tiff/libtiff/tif_fax3.h
@@ -289,39 +289,30 @@ static const char *StateNames[] = {
  * is non-zero then we still need to scan for the final flag
  * bit that is part of the EOL code.
  */
-#define SYNC_EOL(eoflab, retrywithouteol)                                      \
+#define SYNC_EOL(eoflab)                                                       \
     do                                                                         \
     {                                                                          \
-        if (!(sp->b.mode & FAXMODE_NOEOL)) /* skip EOL, if not present */      \
+        if (EOLcnt == 0)                                                       \
         {                                                                      \
-            if (EOLcnt == 0)                                                   \
-            {                                                                  \
-                for (;;)                                                       \
-                {                                                              \
-                    NeedBits16(11, eoflab);                                    \
-                    if (GetBits(11) == 0)                                      \
-                        break; /* EOL found */                                 \
-                    ClrBits(1);                                                \
-                }                                                              \
-            }                                                                  \
-            /* Now move after EOL or detect missing EOL. */                    \
             for (;;)                                                           \
             {                                                                  \
-                NeedBits8(8, noEOLFound);                                      \
-                if (GetBits(8))                                                \
+                NeedBits16(11, eoflab);                                        \
+                if (GetBits(11) == 0)                                          \
                     break;                                                     \
-                ClrBits(8);                                                    \
-            }                                                                  \
-            while (GetBits(1) == 0)                                            \
                 ClrBits(1);                                                    \
-            ClrBits(1); /* EOL bit */                                          \
-            EOLcnt = 0; /* reset EOL counter/flag */                           \
-            break;      /* existing EOL skipped, leave macro */                \
-        noEOLFound:                                                            \
-            sp->b.mode |= FAXMODE_NOEOL;                                       \
-            tryG3WithoutEOL(a0);                                               \
-            goto retrywithouteol;                                              \
+            }                                                                  \
+        }                                                                      \
+        for (;;)                                                               \
+        {                                                                      \
+            NeedBits8(8, eoflab);                                              \
+            if (GetBits(8))                                                    \
+                break;                                                         \
+            ClrBits(8);                                                        \
         }                                                                      \
+        while (GetBits(1) == 0)                                                \
+            ClrBits(1);                                                        \
+        ClrBits(1); /* EOL bit */                                              \
+        EOLcnt = 0; /* reset EOL counter/flag */                               \
     } while (0)
 
 /*
