diff --git a/libs/tiff/libtiff/tif_lzw.c b/libs/tiff/libtiff/tif_lzw.c
index 4bf845b393d6619535c6d6d05d990c2fba5efbc4..4baf78e50b206a1661ffdccc685e034ffb9a2894 100644
--- a/libs/tiff/libtiff/tif_lzw.c
+++ b/libs/tiff/libtiff/tif_lzw.c
@@ -168,6 +168,7 @@ static int LZWDecode(TIFF *tif, uint8_t *op0, tmsize_t occ0, uint16_t s);
 #ifdef LZW_COMPAT
 static int LZWDecodeCompat(TIFF *tif, uint8_t *op0, tmsize_t occ0, uint16_t s);
 #endif
+static void cl_hash(LZWCodecState *);
 
 /*
  * LZW Decoder.
@@ -732,7 +733,6 @@ after_loop:
     if (occ > 0)
     {
         memset(op, 0, (size_t)occ);
-        sp->read_error = 1;
         TIFFErrorExtR(tif, module,
                       "Not enough data at scanline %" PRIu32 " (short %" PRIu64
                       " bytes)",
@@ -1017,10 +1017,6 @@ static int LZWDecodeCompat(TIFF *tif, uint8_t *op0, tmsize_t occ0, uint16_t s)
 }
 #endif /* LZW_COMPAT */
 
-#ifndef LZW_READ_ONLY
-
-static void cl_hash(LZWCodecState *);
-
 /*
  * LZW Encoding.
  */
@@ -1377,13 +1373,11 @@ static void cl_hash(LZWCodecState *sp)
         hp->hash = -1;
 }
 
-#endif
-
 static void LZWCleanup(TIFF *tif)
 {
     (void)TIFFPredictorCleanup(tif);
 
-    assert(tif->tif_data != NULL);
+    assert(tif->tif_data != 0);
 
     if (LZWDecoderState(tif)->dec_codetab)
         _TIFFfreeExt(tif, LZWDecoderState(tif)->dec_codetab);
@@ -1422,14 +1416,12 @@ int TIFFInitLZW(TIFF *tif, int scheme)
     tif->tif_decoderow = LZWDecode;
     tif->tif_decodestrip = LZWDecode;
     tif->tif_decodetile = LZWDecode;
-#ifndef LZW_READ_ONLY
     tif->tif_setupencode = LZWSetupEncode;
     tif->tif_preencode = LZWPreEncode;
     tif->tif_postencode = LZWPostEncode;
     tif->tif_encoderow = LZWEncode;
     tif->tif_encodestrip = LZWEncode;
     tif->tif_encodetile = LZWEncode;
-#endif
     tif->tif_cleanup = LZWCleanup;
     /*
      * Setup predictor setup.
