diff --git a/libs/tiff/libtiff/tif_open.c b/libs/tiff/libtiff/tif_open.c
index 565d13cf6941bc699f5d784a234824af93220e62..59a07bee023230275463a828d26f7e41d58ceb65 100644
--- a/libs/tiff/libtiff/tif_open.c
+++ b/libs/tiff/libtiff/tif_open.c
@@ -37,7 +37,7 @@
 /*
  * Dummy functions to fill the omitted client procedures.
  */
-int _tiffDummyMapProc(thandle_t fd, void **pbase, toff_t *psize)
+static int _tiffDummyMapProc(thandle_t fd, void **pbase, toff_t *psize)
 {
     (void)fd;
     (void)pbase;
@@ -45,7 +45,7 @@ int _tiffDummyMapProc(thandle_t fd, void **pbase, toff_t *psize)
     return (0);
 }
 
-void _tiffDummyUnmapProc(thandle_t fd, void *base, toff_t size)
+static void _tiffDummyUnmapProc(thandle_t fd, void *base, toff_t size)
 {
     (void)fd;
     (void)base;
@@ -109,15 +109,6 @@ void TIFFOpenOptionsSetMaxCumulatedMemAlloc(TIFFOpenOptions *opts,
     opts->max_cumulated_mem_alloc = max_cumulated_mem_alloc;
 }
 
-/** Whether a warning should be emitted when encountering a unknown tag.
- * Default is FALSE since libtiff 4.7.1
- */
-void TIFFOpenOptionsSetWarnAboutUnknownTags(TIFFOpenOptions *opts,
-                                            int warn_about_unknown_tags)
-{
-    opts->warn_about_unknown_tags = warn_about_unknown_tags;
-}
-
 void TIFFOpenOptionsSetErrorHandlerExtR(TIFFOpenOptions *opts,
                                         TIFFErrorHandlerExtR handler,
                                         void *errorhandler_user_data)
@@ -395,7 +386,6 @@ TIFF *TIFFClientOpenExt(const char *name, const char *mode,
         tif->tif_warnhandler_user_data = opts->warnhandler_user_data;
         tif->tif_max_single_mem_alloc = opts->max_single_mem_alloc;
         tif->tif_max_cumulated_mem_alloc = opts->max_cumulated_mem_alloc;
-        tif->tif_warn_about_unknown_tags = opts->warn_about_unknown_tags;
     }
 
     if (!readproc || !writeproc || !seekproc || !closeproc || !sizeproc)
