diff --git a/libs/tiff/libtiff/tif_packbits.c b/libs/tiff/libtiff/tif_packbits.c
index d7db9b64ea3644b47f2802798c742d4026853996..1ae50cbd47ce6975405c0a89a68066f25e0a6462 100644
--- a/libs/tiff/libtiff/tif_packbits.c
+++ b/libs/tiff/libtiff/tif_packbits.c
@@ -31,8 +31,6 @@
  */
 #include <stdio.h>
 
-#ifndef PACKBITS_READ_ONLY
-
 static int PackBitsPreEncode(TIFF *tif, uint16_t s)
 {
     (void)s;
@@ -80,7 +78,7 @@ static int PackBitsEncode(TIFF *tif, uint8_t *buf, tmsize_t cc, uint16_t s)
     op = tif->tif_rawcp;
     ep = tif->tif_rawdata + tif->tif_rawdatasize;
     state = BASE;
-    lastliteral = NULL;
+    lastliteral = 0;
     while (cc > 0)
     {
         /*
@@ -233,8 +231,6 @@ static int PackBitsEncodeChunk(TIFF *tif, uint8_t *bp, tmsize_t cc, uint16_t s)
     return (1);
 }
 
-#endif
-
 static int PackBitsDecode(TIFF *tif, uint8_t *op, tmsize_t occ, uint16_t s)
 {
     static const char module[] = "PackBitsDecode";
@@ -318,13 +314,11 @@ int TIFFInitPackBits(TIFF *tif, int scheme)
     tif->tif_decoderow = PackBitsDecode;
     tif->tif_decodestrip = PackBitsDecode;
     tif->tif_decodetile = PackBitsDecode;
-#ifndef PACKBITS_READ_ONLY
     tif->tif_preencode = PackBitsPreEncode;
     tif->tif_postencode = PackBitsPostEncode;
     tif->tif_encoderow = PackBitsEncode;
     tif->tif_encodestrip = PackBitsEncodeChunk;
     tif->tif_encodetile = PackBitsEncodeChunk;
-#endif
     return (1);
 }
 #endif /* PACKBITS_SUPPORT */
