diff --git a/libs/tiff/libtiff/tif_pixarlog.c b/libs/tiff/libtiff/tif_pixarlog.c
index 309054804d3dab305826928faa752c2d73eb9010..56cf416a7f6b467f9c2bda92a563b54140c28946 100644
--- a/libs/tiff/libtiff/tif_pixarlog.c
+++ b/libs/tiff/libtiff/tif_pixarlog.c
@@ -1596,10 +1596,10 @@ static int PixarLogVGetField(TIFF *tif, uint32_t tag, va_list ap)
 }
 
 static const TIFFField pixarlogFields[] = {
-    {TIFFTAG_PIXARLOGDATAFMT, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT, FIELD_PSEUDO,
-     FALSE, FALSE, "", NULL},
-    {TIFFTAG_PIXARLOGQUALITY, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT, FIELD_PSEUDO,
-     FALSE, FALSE, "", NULL}};
+    {TIFFTAG_PIXARLOGDATAFMT, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT,
+     TIFF_SETGET_UNDEFINED, FIELD_PSEUDO, FALSE, FALSE, "", NULL},
+    {TIFFTAG_PIXARLOGQUALITY, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT,
+     TIFF_SETGET_UNDEFINED, FIELD_PSEUDO, FALSE, FALSE, "", NULL}};
 
 int TIFFInitPixarLog(TIFF *tif, int scheme)
 {
