diff --git a/libs/tiff/libtiff/tif_zip.c b/libs/tiff/libtiff/tif_zip.c
index 8abf88ae7696b6878d82bbb08c40d60aba68f269..2a2a1d787f69b1cddce145fa1675702dc51fffb5 100644
--- a/libs/tiff/libtiff/tif_zip.c
+++ b/libs/tiff/libtiff/tif_zip.c
@@ -128,16 +128,6 @@ static int ZIPSetupDecode(TIFF *tif)
     }
 }
 
-static inline uint64_t TIFF_MIN_UINT64(uint64_t a, uint64_t b)
-{
-    return a < b ? a : b;
-}
-
-static inline uInt TIFF_CLAMP_UINT64_TO_INT32_MAX(uint64_t v)
-{
-    return (uInt)TIFF_MIN_UINT64(v, INT32_MAX);
-}
-
 /*
  * Setup state for decoding a strip.
  */
@@ -159,7 +149,9 @@ static int ZIPPreDecode(TIFF *tif, uint16_t s)
          we need to simplify this code to reflect a ZLib that is likely updated
          to deal with 8byte memory sizes, though this code will respond
          appropriately even before we simplify it */
-    sp->stream.avail_in = TIFF_CLAMP_UINT64_TO_INT32_MAX(tif->tif_rawcc);
+    sp->stream.avail_in = (uint64_t)tif->tif_rawcc < 0xFFFFFFFFU
+                              ? (uInt)tif->tif_rawcc
+                              : 0xFFFFFFFFU;
     if (inflateReset(&sp->stream) == Z_OK)
     {
         sp->read_error = 0;
@@ -275,10 +267,14 @@ static int ZIPDecode(TIFF *tif, uint8_t *op, tmsize_t occ, uint16_t s)
     do
     {
         int state;
-        uInt avail_in_before = TIFF_CLAMP_UINT64_TO_INT32_MAX(tif->tif_rawcc);
-        uInt avail_out_before = TIFF_CLAMP_UINT64_TO_INT32_MAX(occ);
+        uInt avail_in_before = (uint64_t)tif->tif_rawcc <= 0xFFFFFFFFU
+                                   ? (uInt)tif->tif_rawcc
+                                   : 0xFFFFFFFFU;
+        uInt avail_out_before =
+            (uint64_t)occ < 0xFFFFFFFFU ? (uInt)occ : 0xFFFFFFFFU;
         sp->stream.avail_in = avail_in_before;
         sp->stream.avail_out = avail_out_before;
+        /* coverity[overrun-buffer-arg] */
         state = inflate(&sp->stream, Z_PARTIAL_FLUSH);
         tif->tif_rawcc -= (avail_in_before - sp->stream.avail_in);
         occ -= (avail_out_before - sp->stream.avail_out);
@@ -286,7 +282,7 @@ static int ZIPDecode(TIFF *tif, uint8_t *op, tmsize_t occ, uint16_t s)
             break;
         if (state == Z_DATA_ERROR)
         {
-            memset(sp->stream.next_out, 0, (size_t)occ);
+            memset(sp->stream.next_out, 0, sp->stream.avail_out);
             TIFFErrorExtR(tif, module, "Decoding error at scanline %lu, %s",
                           (unsigned long)tif->tif_row, SAFE_MSG(sp));
             sp->read_error = 1;
@@ -294,7 +290,7 @@ static int ZIPDecode(TIFF *tif, uint8_t *op, tmsize_t occ, uint16_t s)
         }
         if (state != Z_OK)
         {
-            memset(sp->stream.next_out, 0, (size_t)occ);
+            memset(sp->stream.next_out, 0, sp->stream.avail_out);
             TIFFErrorExtR(tif, module, "ZLib error: %s", SAFE_MSG(sp));
             sp->read_error = 1;
             return (0);
@@ -306,7 +302,7 @@ static int ZIPDecode(TIFF *tif, uint8_t *op, tmsize_t occ, uint16_t s)
                       "Not enough data at scanline %lu (short %" PRIu64
                       " bytes)",
                       (unsigned long)tif->tif_row, (uint64_t)occ);
-        memset(sp->stream.next_out, 0, (size_t)occ);
+        memset(sp->stream.next_out, 0, sp->stream.avail_out);
         sp->read_error = 1;
         return (0);
     }
@@ -485,8 +481,10 @@ static int ZIPEncode(TIFF *tif, uint8_t *bp, tmsize_t cc, uint16_t s)
          appropriately even before we simplify it */
     do
     {
-        uInt avail_in_before = TIFF_CLAMP_UINT64_TO_INT32_MAX(cc);
+        uInt avail_in_before =
+            (uint64_t)cc <= 0xFFFFFFFFU ? (uInt)cc : 0xFFFFFFFFU;
         sp->stream.avail_in = avail_in_before;
+        /* coverity[overrun-buffer-arg] */
         if (deflate(&sp->stream, Z_NO_FLUSH) != Z_OK)
         {
             TIFFErrorExtR(tif, module, "Encoder error: %s", SAFE_MSG(sp));
@@ -498,8 +496,9 @@ static int ZIPEncode(TIFF *tif, uint8_t *bp, tmsize_t cc, uint16_t s)
             if (!TIFFFlushData1(tif))
                 return 0;
             sp->stream.next_out = tif->tif_rawdata;
-            sp->stream.avail_out =
-                TIFF_CLAMP_UINT64_TO_INT32_MAX(tif->tif_rawdatasize);
+            sp->stream.avail_out = (uint64_t)tif->tif_rawdatasize <= 0xFFFFFFFFU
+                                       ? (uInt)tif->tif_rawdatasize
+                                       : 0xFFFFFFFFU;
         }
         cc -= (avail_in_before - sp->stream.avail_in);
     } while (cc > 0);
@@ -673,31 +672,12 @@ static int ZIPVGetField(TIFF *tif, uint32_t tag, va_list ap)
 }
 
 static const TIFFField zipFields[] = {
-    {TIFFTAG_ZIPQUALITY, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT, FIELD_PSEUDO, TRUE,
-     FALSE, "", NULL},
-    {TIFFTAG_DEFLATE_SUBCODEC, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT, FIELD_PSEUDO,
-     TRUE, FALSE, "", NULL},
+    {TIFFTAG_ZIPQUALITY, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT,
+     TIFF_SETGET_UNDEFINED, FIELD_PSEUDO, TRUE, FALSE, "", NULL},
+    {TIFFTAG_DEFLATE_SUBCODEC, 0, 0, TIFF_ANY, 0, TIFF_SETGET_INT,
+     TIFF_SETGET_UNDEFINED, FIELD_PSEUDO, TRUE, FALSE, "", NULL},
 };
 
-static void *TIFF_zalloc(void *opaque, unsigned int items, unsigned int size)
-{
-    static const char module[] = "TIFF_zalloc";
-    TIFF *tif = opaque;
-
-    if (items > ~(size_t)0 / size)
-    {
-        TIFFErrorExtR(tif, module, "Overflow");
-        return NULL;
-    }
-
-    return _TIFFmallocExt(tif, items * size);
-}
-
-static void TIFF_zfree(void *opaque, void *ptr)
-{
-    _TIFFfreeExt((TIFF *)opaque, ptr);
-}
-
 int TIFFInitZIP(TIFF *tif, int scheme)
 {
     static const char module[] = "TIFFInitZIP";
@@ -726,9 +706,9 @@ int TIFFInitZIP(TIFF *tif, int scheme)
     if (tif->tif_data == NULL)
         goto bad;
     sp = GetZIPState(tif);
-    sp->stream.zalloc = TIFF_zalloc;
-    sp->stream.zfree = TIFF_zfree;
-    sp->stream.opaque = tif;
+    sp->stream.zalloc = NULL;
+    sp->stream.zfree = NULL;
+    sp->stream.opaque = NULL;
     sp->stream.data_type = Z_BINARY;
 
     /*
