diff --git a/libs/tiff/libtiff/tiffconf.h b/libs/tiff/libtiff/tiffconf.h
index 988ccceb46ec1e8ebe7e9b1de2b1349db68de0cc..b5a667c8edf0d817ecc5f0944b83c486b7e0d022 100644
--- a/libs/tiff/libtiff/tiffconf.h
+++ b/libs/tiff/libtiff/tiffconf.h
@@ -73,7 +73,7 @@
 #define CCITT_SUPPORT 1
 
 /* Support JPEG compression (requires IJG JPEG library) */
-#define JPEG_SUPPORT
+/* #undef JPEG_SUPPORT */
 
 /* Support JBIG compression (requires JBIG-KIT library) */
 /* #undef JBIG_SUPPORT */
@@ -90,7 +90,7 @@
 /* Support NeXT 2-bit RLE algorithm */
 #define NEXT_SUPPORT 1
 
-/* Support Old JPEG compression (read contrib/ojpeg/README first! Compilation
+/* Support Old JPEG compresson (read contrib/ojpeg/README first! Compilation
    fails with unpatched IJG JPEG library) */
 /* #undef OJPEG_SUPPORT */
 
