diff --git a/libs/tiff/libtiff/tiffio.h b/libs/tiff/libtiff/tiffio.h
index f9c206e32a8b7a7767d65561791257df85c3a8a8..225f3c1bfeabae2b31a5f35fbfd12d14b570dcd8 100644
--- a/libs/tiff/libtiff/tiffio.h
+++ b/libs/tiff/libtiff/tiffio.h
@@ -45,14 +45,12 @@ typedef struct tiff TIFF;
  * to pass tag types and values uses the types defined in
  * tiff.h directly.
  *
- * NB: ttag_t -> deprecated and replaced by uint32_t
- *     is unsigned int and not unsigned short because
+ * NB: ttag_t is unsigned int and not unsigned short because
  *     ANSI C requires that the type before the ellipsis be a
  *     promoted type (i.e. one of int, unsigned int, pointer,
  *     or double) and because we defined pseudo-tags that are
  *     outside the range of legal Aldus-assigned tags.
- * NB: tsize_t -> deprecated and replaced by tmsize_t
- *     is signed and not unsigned because some functions
+ * NB: tsize_t is signed and not unsigned because some functions
  *     return -1.
  * NB: toff_t is not off_t for many reasons; TIFFs max out at
  *     32-bit file offsets, and BigTIFF maxes out at 64-bit
@@ -68,11 +66,10 @@ typedef TIFF_SSIZE_T tmsize_t;
 #define TIFF_TMSIZE_T_MAX (tmsize_t)(SIZE_MAX >> 1)
 
 typedef uint64_t toff_t; /* file offset */
-typedef uint32_t tdir_t; /* directory index */
-
 /* the following are deprecated and should be replaced by their defining
    counterparts */
 typedef uint32_t ttag_t;    /* directory tag */
+typedef uint32_t tdir_t;    /* directory index */
 typedef uint16_t tsample_t; /* sample number */
 typedef uint32_t tstrile_t; /* strip or tile number */
 typedef tstrile_t tstrip_t; /* strip number */
@@ -154,18 +151,6 @@ typedef struct
     float d_gammaB;
 } TIFFDisplay;
 
-/* YCbCr->RGB support for TIFFYCbCrToRGBInit() and TIFFYCbCrToRGB()
- * Attention:
- * Functions TIFFYCbCrToRGBInit() and TIFFYCbCrToRGB() require a user provided
- * large memory buffer, where several tables can be setup.
- * The pointers to these tables are stored in the structure TIFFYCbCrToRGB,
- * which is located at the beginning of the buffer. Thus, this memory has to be
- * allocated as follows:
- *     TIFFYCbCrToRGB *ycbcr = (TIFFYCbCrToRGB *)_TIFFmalloc(
- *         TIFFroundup_32(sizeof(TIFFYCbCrToRGB), sizeof(long)) +
- *         4 * 256 * sizeof(TIFFRGBValue) + 2 * 256 * sizeof(int) +
- *         3 * 256 * sizeof(int32_t));
- */
 typedef struct
 {                           /* YCbCr->RGB support */
     TIFFRGBValue *clamptab; /* range clamping table */
@@ -523,9 +508,6 @@ extern int TIFFReadRGBAImageOriented(TIFF *, uint32_t, uint32_t, uint32_t *,
     TIFFOpenOptionsSetMaxCumulatedMemAlloc(TIFFOpenOptions *opts,
                                            tmsize_t max_cumulated_mem_alloc);
     extern void
-    TIFFOpenOptionsSetWarnAboutUnknownTags(TIFFOpenOptions *opts,
-                                           int warn_about_unknown_tags);
-    extern void
     TIFFOpenOptionsSetErrorHandlerExtR(TIFFOpenOptions *opts,
                                        TIFFErrorHandlerExtR handler,
                                        void *errorhandler_user_data);
@@ -608,7 +590,7 @@ extern int TIFFReadRGBAImageOriented(TIFF *, uint32_t, uint32_t, uint32_t *,
     extern uint64_t TIFFGetStrileByteCountWithErr(TIFF *tif, uint32_t strile,
                                                   int *pbErr);
 
-#if LOGLUV_PUBLIC
+#ifdef LOGLUV_PUBLIC
 #define U_NEU 0.210526316
 #define V_NEU 0.473684211
 #define UVSCALE 410.
@@ -652,7 +634,7 @@ extern int TIFFReadRGBAImageOriented(TIFF *, uint32_t, uint32_t, uint32_t *,
      ****************************************************************************/
     typedef struct
     {
-        uint32_t field_tag;             /* field's tag */
+        ttag_t field_tag;               /* field's tag */
         short field_readcount;          /* read count/TIFF_VARIABLE/TIFF_SPP */
         short field_writecount;         /* write count/TIFF_VARIABLE */
         TIFFDataType field_type;        /* type of associated data */
