diff --git a/libs/tiff/libtiff/tiffiop.h b/libs/tiff/libtiff/tiffiop.h
index 80c04c279be4f917302eb9bb92cd85829db13da0..c4348206df5670e2f9c97a111e8bd620c9992b4a 100644
--- a/libs/tiff/libtiff/tiffiop.h
+++ b/libs/tiff/libtiff/tiffiop.h
@@ -258,7 +258,6 @@ struct tiff
     tmsize_t tif_max_single_mem_alloc;    /* in bytes. 0 for unlimited */
     tmsize_t tif_max_cumulated_mem_alloc; /* in bytes. 0 for unlimited */
     tmsize_t tif_cur_cumulated_mem_alloc; /* in bytes */
-    int tif_warn_about_unknown_tags;
 };
 
 struct TIFFOpenOptions
@@ -269,7 +268,6 @@ struct TIFFOpenOptions
     void *warnhandler_user_data;       /* may be NULL */
     tmsize_t max_single_mem_alloc;     /* in bytes. 0 for unlimited */
     tmsize_t max_cumulated_mem_alloc;  /* in bytes. 0 for unlimited */
-    int warn_about_unknown_tags;
 };
 
 #define isPseudoTag(t) (t > 0xffff) /* is tag value normal or pseudo */
@@ -400,8 +398,6 @@ typedef size_t TIFFIOSize_t;
 extern "C"
 {
 #endif
-    extern int _tiffDummyMapProc(thandle_t fd, void **pbase, toff_t *psize);
-    extern void _tiffDummyUnmapProc(thandle_t fd, void *base, toff_t size);
     extern int _TIFFgetMode(TIFFOpenOptions *opts, thandle_t clientdata,
                             const char *mode, const char *module);
     extern int _TIFFNoRowEncode(TIFF *tif, uint8_t *pp, tmsize_t cc,
