diff --git a/libs/vkd3d/include/private/vkd3d_common.h b/libs/vkd3d/include/private/vkd3d_common.h
index 7ee11b54396d6838e7fc66b7e604b9cbf3cc1b0e..beb23257fb77d13eb42c9db06a8599f7ce820ad3 100644
--- a/libs/vkd3d/include/private/vkd3d_common.h
+++ b/libs/vkd3d/include/private/vkd3d_common.h
@@ -19,11 +19,6 @@
 #ifndef __VKD3D_COMMON_H
 #define __VKD3D_COMMON_H
 
-#ifndef VKD3D_DEBUG_ENV
-#define VKD3D_DEBUG_ENV VKD3D_DEBUG
-#endif
-#define VKD3D_DEBUG_ENV_NAME VKD3D_EXPAND_AND_STRINGIFY(VKD3D_DEBUG_ENV)
-
 #include "config.h"
 #define WIN32_LEAN_AND_MEAN
 #include "windows.h"
@@ -43,8 +38,6 @@
 #include <intrin.h>
 #endif
 
-#define VKD3D_SHADER_API_VERSION_CURRENT VKD3D_SHADER_API_VERSION_1_18
-
 #ifndef ARRAY_SIZE
 # define ARRAY_SIZE(x) (sizeof(x) / sizeof(*(x)))
 #endif
@@ -73,7 +66,6 @@
 
 #define TAG_AON9 VKD3D_MAKE_TAG('A', 'o', 'n', '9')
 #define TAG_CLI4 VKD3D_MAKE_TAG('C', 'L', 'I', '4')
-#define TAG_CLIT VKD3D_MAKE_TAG('C', 'L', 'I', 'T')
 #define TAG_CTAB VKD3D_MAKE_TAG('C', 'T', 'A', 'B')
 #define TAG_DXBC VKD3D_MAKE_TAG('D', 'X', 'B', 'C')
 #define TAG_DXIL VKD3D_MAKE_TAG('D', 'X', 'I', 'L')
@@ -154,10 +146,9 @@ enum vkd3d_dbg_level
     VKD3D_DBG_LEVEL_TRACE,
 };
 
-enum vkd3d_dbg_level vkd3d_dbg_get_level(const char *vkd3d_dbg_env_name);
+enum vkd3d_dbg_level vkd3d_dbg_get_level(void);
 
-void vkd3d_dbg_printf(const char *vkd3d_dbg_env_name, enum vkd3d_dbg_level level,
-        const char *function, const char *fmt, ...) VKD3D_PRINTF_FUNC(4, 5);
+void vkd3d_dbg_printf(enum vkd3d_dbg_level level, const char *function, const char *fmt, ...) VKD3D_PRINTF_FUNC(3, 4);
 void vkd3d_dbg_set_log_callback(PFN_vkd3d_log callback);
 
 const char *vkd3d_dbg_sprintf(const char *fmt, ...) VKD3D_PRINTF_FUNC(1, 2);
@@ -180,7 +171,7 @@ const char *debugstr_w(const WCHAR *wstr, size_t wchar_size);
         VKD3D_DBG_PRINTF_##level
 
 #define VKD3D_DBG_PRINTF(...) \
-        vkd3d_dbg_printf(VKD3D_DEBUG_ENV_NAME, vkd3d_dbg_level, __FUNCTION__, __VA_ARGS__); } while (0)
+        vkd3d_dbg_printf(vkd3d_dbg_level, __FUNCTION__, __VA_ARGS__); } while (0)
 
 #define VKD3D_DBG_PRINTF_TRACE(...) VKD3D_DBG_PRINTF(__VA_ARGS__)
 #define VKD3D_DBG_PRINTF_WARN(...) VKD3D_DBG_PRINTF(__VA_ARGS__)
@@ -189,7 +180,7 @@ const char *debugstr_w(const WCHAR *wstr, size_t wchar_size);
 
 #ifdef VKD3D_ABORT_ON_ERR
 #define VKD3D_DBG_PRINTF_ERR(...) \
-        vkd3d_dbg_printf(VKD3D_DEBUG_ENV_NAME, vkd3d_dbg_level, __FUNCTION__, __VA_ARGS__); \
+        vkd3d_dbg_printf(vkd3d_dbg_level, __FUNCTION__, __VA_ARGS__); \
         abort(); \
         } while (0)
 #else
@@ -223,17 +214,19 @@ const char *debugstr_w(const WCHAR *wstr, size_t wchar_size);
 #endif
 
 #ifndef TRACE_ON
-#define TRACE_ON() (vkd3d_dbg_get_level(VKD3D_DEBUG_ENV_NAME) == VKD3D_DBG_LEVEL_TRACE)
+#define TRACE_ON() (vkd3d_dbg_get_level() == VKD3D_DBG_LEVEL_TRACE)
 #endif
 
 #ifndef WARN_ON
-#define WARN_ON() (vkd3d_dbg_get_level(VKD3D_DEBUG_ENV_NAME) >= VKD3D_DBG_LEVEL_WARN)
+#define WARN_ON() (vkd3d_dbg_get_level() >= VKD3D_DBG_LEVEL_WARN)
 #endif
 
 #ifndef FIXME_ONCE
 #define FIXME_ONCE VKD3D_DBG_LOG_ONCE(FIXME, WARN)
 #endif
 
+#define VKD3D_DEBUG_ENV_NAME(name) const char *const vkd3d_dbg_env_name = name
+
 static inline const char *debugstr_guid(const GUID *guid)
 {
     if (!guid)
@@ -347,29 +340,6 @@ static inline unsigned int vkd3d_log2i(unsigned int x)
 #endif
 }
 
-static inline unsigned int vkd3d_ctz(uint32_t v)
-{
-#ifdef _WIN32
-    ULONG result;
-    if (_BitScanForward(&result, v))
-        return (unsigned int)result;
-    return 32;
-#elif defined(HAVE_BUILTIN_CTZ)
-    return __builtin_ctz(v);
-#else
-    unsigned int c = 31;
-
-    v &= -v;
-    c = (v & 0x0000ffff) ? c - 16 : c;
-    c = (v & 0x00ff00ff) ? c - 8 : c;
-    c = (v & 0x0f0f0f0f) ? c - 4 : c;
-    c = (v & 0x33333333) ? c - 2 : c;
-    c = (v & 0x55555555) ? c - 1 : c;
-
-    return c;
-#endif
-}
-
 static inline void *vkd3d_memmem( const void *haystack, size_t haystack_len, const void *needle, size_t needle_len)
 {
     const char *str = haystack;
