diff --git a/libs/vkd3d/libs/vkd3d-common/debug.c b/libs/vkd3d/libs/vkd3d-common/debug.c
index b5d74ca3dc7a6e2c37d0b0082a68a326decf689c..32862024b90666d50c1cd08fdc272dfe83f4d02b 100644
--- a/libs/vkd3d/libs/vkd3d-common/debug.c
+++ b/libs/vkd3d/libs/vkd3d-common/debug.c
@@ -40,6 +40,8 @@
 #define VKD3D_DEBUG_BUFFER_COUNT 64
 #define VKD3D_DEBUG_BUFFER_SIZE 512
 
+extern const char *const vkd3d_dbg_env_name;
+
 static const char *const debug_level_names[] =
 {
     [VKD3D_DBG_LEVEL_NONE ] =   "none",
@@ -50,7 +52,7 @@ static const char *const debug_level_names[] =
     [VKD3D_DBG_LEVEL_TRACE] =   "trace",
 };
 
-enum vkd3d_dbg_level vkd3d_dbg_get_level(const char *vkd3d_dbg_env_name)
+enum vkd3d_dbg_level vkd3d_dbg_get_level(void)
 {
     static unsigned int level = ~0u;
     const char *vkd3d_debug;
@@ -106,12 +108,11 @@ static uint64_t get_pthread_threadid(void)
 }
 #endif
 
-void vkd3d_dbg_printf(const char *vkd3d_dbg_env_name,
-        enum vkd3d_dbg_level level, const char *function, const char *fmt, ...)
+void vkd3d_dbg_printf(enum vkd3d_dbg_level level, const char *function, const char *fmt, ...)
 {
     va_list args;
 
-    if (vkd3d_dbg_get_level(vkd3d_dbg_env_name) < level)
+    if (vkd3d_dbg_get_level() < level)
         return;
 
 #ifdef _WIN32
