diff --git a/libs/vkd3d/libs/vkd3d-shader/vkd3d_shader_private.h b/libs/vkd3d/libs/vkd3d-shader/vkd3d_shader_private.h
index 97fe5238046021c57175943580c3950b4b9573fd..3b4fb626fcc9fc5987341cdcf4b9575a69c7e30e 100644
--- a/libs/vkd3d/libs/vkd3d-shader/vkd3d_shader_private.h
+++ b/libs/vkd3d/libs/vkd3d-shader/vkd3d_shader_private.h
@@ -110,7 +110,6 @@ enum vkd3d_shader_error
     VKD3D_SHADER_WARNING_SPV_INVALID_SWIZZLE            = 2300,
     VKD3D_SHADER_WARNING_SPV_INVALID_UAV_FLAGS          = 2301,
     VKD3D_SHADER_WARNING_SPV_IGNORING_FLAG              = 2302,
-    VKD3D_SHADER_WARNING_SPV_INVALID_SIZE               = 2303,
 
     VKD3D_SHADER_ERROR_RS_OUT_OF_MEMORY                 = 3000,
     VKD3D_SHADER_ERROR_RS_INVALID_VERSION               = 3001,
@@ -175,8 +174,6 @@ enum vkd3d_shader_error
     VKD3D_SHADER_ERROR_HLSL_INVALID_MAX_VERTEX_COUNT    = 5042,
     VKD3D_SHADER_ERROR_HLSL_MISSING_PRIMITIVE_TYPE      = 5043,
     VKD3D_SHADER_ERROR_HLSL_MISPLACED_STREAM_OUTPUT     = 5044,
-    VKD3D_SHADER_ERROR_HLSL_MISSING_INPUT_PATCH         = 5045,
-    VKD3D_SHADER_ERROR_HLSL_CANNOT_FLATTEN              = 5046,
 
     VKD3D_SHADER_WARNING_HLSL_IMPLICIT_TRUNCATION       = 5300,
     VKD3D_SHADER_WARNING_HLSL_DIVISION_BY_ZERO          = 5301,
@@ -185,8 +182,6 @@ enum vkd3d_shader_error
     VKD3D_SHADER_WARNING_HLSL_NON_FINITE_RESULT         = 5304,
     VKD3D_SHADER_WARNING_HLSL_IGNORED_ATTRIBUTE         = 5305,
     VKD3D_SHADER_WARNING_HLSL_IGNORED_DEFAULT_VALUE     = 5306,
-    VKD3D_SHADER_WARNING_HLSL_IGNORED_MODIFIER          = 5307,
-    VKD3D_SHADER_WARNING_HLSL_OVERRIDDEN_SEMANTIC       = 5308,
 
     VKD3D_SHADER_ERROR_GLSL_INTERNAL                    = 6000,
     VKD3D_SHADER_ERROR_GLSL_BINDING_NOT_FOUND           = 6001,
@@ -215,7 +210,7 @@ enum vkd3d_shader_error
     VKD3D_SHADER_ERROR_DXIL_INVALID_OPERAND_COUNT       = 8005,
     VKD3D_SHADER_ERROR_DXIL_INVALID_TYPE_TABLE          = 8006,
     VKD3D_SHADER_ERROR_DXIL_INVALID_VALUE_SYMTAB        = 8007,
-    VKD3D_SHADER_ERROR_DXIL_UNSUPPORTED                 = 8008,
+    VKD3D_SHADER_ERROR_DXIL_UNSUPPORTED_BITCODE_FORMAT  = 8008,
     VKD3D_SHADER_ERROR_DXIL_INVALID_FUNCTION_DCL        = 8009,
     VKD3D_SHADER_ERROR_DXIL_INVALID_TYPE_ID             = 8010,
     VKD3D_SHADER_ERROR_DXIL_INVALID_MODULE              = 8011,
@@ -227,8 +222,6 @@ enum vkd3d_shader_error
     VKD3D_SHADER_ERROR_DXIL_INVALID_PROPERTIES          = 8017,
     VKD3D_SHADER_ERROR_DXIL_INVALID_RESOURCES           = 8018,
     VKD3D_SHADER_ERROR_DXIL_INVALID_RESOURCE_HANDLE     = 8019,
-    VKD3D_SHADER_ERROR_DXIL_INVALID_CONSTANT            = 8020,
-    VKD3D_SHADER_ERROR_DXIL_NOT_IMPLEMENTED             = 8021,
 
     VKD3D_SHADER_WARNING_DXIL_UNKNOWN_MAGIC_NUMBER      = 8300,
     VKD3D_SHADER_WARNING_DXIL_UNKNOWN_SHADER_TYPE       = 8301,
@@ -243,7 +236,7 @@ enum vkd3d_shader_error
     VKD3D_SHADER_WARNING_DXIL_UNDEFINED_OPERAND         = 8310,
 
     VKD3D_SHADER_ERROR_VSIR_NOT_IMPLEMENTED             = 9000,
-    VKD3D_SHADER_ERROR_VSIR_INVALID_OPCODE              = 9001,
+    VKD3D_SHADER_ERROR_VSIR_INVALID_HANDLER             = 9001,
     VKD3D_SHADER_ERROR_VSIR_INVALID_REGISTER_TYPE       = 9002,
     VKD3D_SHADER_ERROR_VSIR_INVALID_WRITE_MASK          = 9003,
     VKD3D_SHADER_ERROR_VSIR_INVALID_MODIFIERS           = 9004,
@@ -269,7 +262,6 @@ enum vkd3d_shader_error
     VKD3D_SHADER_ERROR_VSIR_INVALID_DESCRIPTOR_TYPE     = 9024,
     VKD3D_SHADER_ERROR_VSIR_INVALID_RESOURCE_TYPE       = 9025,
     VKD3D_SHADER_ERROR_VSIR_INVALID_DESCRIPTOR_COUNT    = 9026,
-    VKD3D_SHADER_ERROR_VSIR_INVALID_FLAGS               = 9027,
 
     VKD3D_SHADER_WARNING_VSIR_DYNAMIC_DESCRIPTOR_ARRAY  = 9300,
 
@@ -286,347 +278,334 @@ enum vkd3d_shader_error
 
 enum vkd3d_shader_opcode
 {
-    VSIR_OP_ABS,
-    VSIR_OP_ACOS,
-    VSIR_OP_ADD,
-    VSIR_OP_AND,
-    VSIR_OP_ASIN,
-    VSIR_OP_ATAN,
-    VSIR_OP_ATOMIC_AND,
-    VSIR_OP_ATOMIC_CMP_STORE,
-    VSIR_OP_ATOMIC_IADD,
-    VSIR_OP_ATOMIC_IMAX,
-    VSIR_OP_ATOMIC_IMIN,
-    VSIR_OP_ATOMIC_OR,
-    VSIR_OP_ATOMIC_UMAX,
-    VSIR_OP_ATOMIC_UMIN,
-    VSIR_OP_ATOMIC_XOR,
-    VSIR_OP_BEM,
-    VSIR_OP_BFI,
-    VSIR_OP_BFREV,
-    VSIR_OP_BRANCH,
-    VSIR_OP_BREAK,
-    VSIR_OP_BREAKC,
-    VSIR_OP_BREAKP,
-    VSIR_OP_BUFINFO,
-    VSIR_OP_CALL,
-    VSIR_OP_CALLNZ,
-    VSIR_OP_CASE,
-    VSIR_OP_CHECK_ACCESS_FULLY_MAPPED,
-    VSIR_OP_CMP,
-    VSIR_OP_CND,
-    VSIR_OP_CONTINUE,
-    VSIR_OP_CONTINUEP,
-    VSIR_OP_COS,
-    VSIR_OP_COUNTBITS,
-    VSIR_OP_CRS,
-    VSIR_OP_CUT,
-    VSIR_OP_CUT_STREAM,
-    VSIR_OP_DADD,
-    VSIR_OP_DCL,
-    VSIR_OP_DCL_CONSTANT_BUFFER,
-    VSIR_OP_DCL_FUNCTION_BODY,
-    VSIR_OP_DCL_FUNCTION_TABLE,
-    VSIR_OP_DCL_GLOBAL_FLAGS,
-    VSIR_OP_DCL_GS_INSTANCES,
-    VSIR_OP_DCL_HS_FORK_PHASE_INSTANCE_COUNT,
-    VSIR_OP_DCL_HS_JOIN_PHASE_INSTANCE_COUNT,
-    VSIR_OP_DCL_HS_MAX_TESSFACTOR,
-    VSIR_OP_DCL_IMMEDIATE_CONSTANT_BUFFER,
-    VSIR_OP_DCL_INDEX_RANGE,
-    VSIR_OP_DCL_INDEXABLE_TEMP,
-    VSIR_OP_DCL_INPUT,
-    VSIR_OP_DCL_INPUT_CONTROL_POINT_COUNT,
-    VSIR_OP_DCL_INPUT_PRIMITIVE,
-    VSIR_OP_DCL_INPUT_PS,
-    VSIR_OP_DCL_INPUT_PS_SGV,
-    VSIR_OP_DCL_INPUT_PS_SIV,
-    VSIR_OP_DCL_INPUT_SGV,
-    VSIR_OP_DCL_INPUT_SIV,
-    VSIR_OP_DCL_INTERFACE,
-    VSIR_OP_DCL_OUTPUT,
-    VSIR_OP_DCL_OUTPUT_CONTROL_POINT_COUNT,
-    VSIR_OP_DCL_OUTPUT_SGV,
-    VSIR_OP_DCL_OUTPUT_SIV,
-    VSIR_OP_DCL_OUTPUT_TOPOLOGY,
-    VSIR_OP_DCL_RESOURCE_RAW,
-    VSIR_OP_DCL_RESOURCE_STRUCTURED,
-    VSIR_OP_DCL_SAMPLER,
-    VSIR_OP_DCL_STREAM,
-    VSIR_OP_DCL_TEMPS,
-    VSIR_OP_DCL_TESSELLATOR_DOMAIN,
-    VSIR_OP_DCL_TESSELLATOR_OUTPUT_PRIMITIVE,
-    VSIR_OP_DCL_TESSELLATOR_PARTITIONING,
-    VSIR_OP_DCL_TGSM_RAW,
-    VSIR_OP_DCL_TGSM_STRUCTURED,
-    VSIR_OP_DCL_THREAD_GROUP,
-    VSIR_OP_DCL_UAV_RAW,
-    VSIR_OP_DCL_UAV_STRUCTURED,
-    VSIR_OP_DCL_UAV_TYPED,
-    VSIR_OP_DCL_VERTICES_OUT,
-    VSIR_OP_DDIV,
-    VSIR_OP_DEF,
-    VSIR_OP_DEFAULT,
-    VSIR_OP_DEFB,
-    VSIR_OP_DEFI,
-    VSIR_OP_DEQO,
-    VSIR_OP_DFMA,
-    VSIR_OP_DGEO,
-    VSIR_OP_DISCARD,
-    VSIR_OP_DIV,
-    VSIR_OP_DLT,
-    VSIR_OP_DMAX,
-    VSIR_OP_DMIN,
-    VSIR_OP_DMOV,
-    VSIR_OP_DMOVC,
-    VSIR_OP_DMUL,
-    VSIR_OP_DNE,
-    VSIR_OP_DP2,
-    VSIR_OP_DP2ADD,
-    VSIR_OP_DP3,
-    VSIR_OP_DP4,
-    VSIR_OP_DRCP,
-    VSIR_OP_DST,
-    VSIR_OP_DSX,
-    VSIR_OP_DSX_COARSE,
-    VSIR_OP_DSX_FINE,
-    VSIR_OP_DSY,
-    VSIR_OP_DSY_COARSE,
-    VSIR_OP_DSY_FINE,
-    VSIR_OP_DTOF,
-    VSIR_OP_DTOI,
-    VSIR_OP_DTOU,
-    VSIR_OP_ELSE,
-    VSIR_OP_EMIT,
-    VSIR_OP_EMIT_STREAM,
-    VSIR_OP_ENDIF,
-    VSIR_OP_ENDLOOP,
-    VSIR_OP_ENDREP,
-    VSIR_OP_ENDSWITCH,
-    VSIR_OP_EQO,
-    VSIR_OP_EQU,
-    VSIR_OP_EVAL_CENTROID,
-    VSIR_OP_EVAL_SAMPLE_INDEX,
-    VSIR_OP_EXP,
-    VSIR_OP_EXPP,
-    VSIR_OP_F16TOF32,
-    VSIR_OP_F32TOF16,
-    VSIR_OP_FCALL,
-    VSIR_OP_FIRSTBIT_HI,
-    VSIR_OP_FIRSTBIT_LO,
-    VSIR_OP_FIRSTBIT_SHI,
-    VSIR_OP_FRC,
-    VSIR_OP_FREM,
-    VSIR_OP_FTOD,
-    VSIR_OP_FTOI,
-    VSIR_OP_FTOU,
-    VSIR_OP_GATHER4,
-    VSIR_OP_GATHER4_C,
-    VSIR_OP_GATHER4_C_S,
-    VSIR_OP_GATHER4_PO,
-    VSIR_OP_GATHER4_PO_C,
-    VSIR_OP_GATHER4_PO_C_S,
-    VSIR_OP_GATHER4_PO_S,
-    VSIR_OP_GATHER4_S,
-    VSIR_OP_GEO,
-    VSIR_OP_GEU,
-    VSIR_OP_HCOS,
-    VSIR_OP_HS_CONTROL_POINT_PHASE,
-    VSIR_OP_HS_DECLS,
-    VSIR_OP_HS_FORK_PHASE,
-    VSIR_OP_HS_JOIN_PHASE,
-    VSIR_OP_HSIN,
-    VSIR_OP_HTAN,
-    VSIR_OP_IADD,
-    VSIR_OP_IBFE,
-    VSIR_OP_IDIV,
-    VSIR_OP_IEQ,
-    VSIR_OP_IF,
-    VSIR_OP_IFC,
-    VSIR_OP_IGE,
-    VSIR_OP_ILT,
-    VSIR_OP_IMAD,
-    VSIR_OP_IMAX,
-    VSIR_OP_IMIN,
-    VSIR_OP_IMM_ATOMIC_ALLOC,
-    VSIR_OP_IMM_ATOMIC_AND,
-    VSIR_OP_IMM_ATOMIC_CMP_EXCH,
-    VSIR_OP_IMM_ATOMIC_CONSUME,
-    VSIR_OP_IMM_ATOMIC_EXCH,
-    VSIR_OP_IMM_ATOMIC_IADD,
-    VSIR_OP_IMM_ATOMIC_IMAX,
-    VSIR_OP_IMM_ATOMIC_IMIN,
-    VSIR_OP_IMM_ATOMIC_OR,
-    VSIR_OP_IMM_ATOMIC_UMAX,
-    VSIR_OP_IMM_ATOMIC_UMIN,
-    VSIR_OP_IMM_ATOMIC_XOR,
-    VSIR_OP_IMUL,
-    VSIR_OP_IMUL_LOW,
-    VSIR_OP_INE,
-    VSIR_OP_INEG,
-    VSIR_OP_IREM,
-    VSIR_OP_ISFINITE,
-    VSIR_OP_ISHL,
-    VSIR_OP_ISHR,
-    VSIR_OP_ISINF,
-    VSIR_OP_ISNAN,
-    VSIR_OP_ITOD,
-    VSIR_OP_ITOF,
-    VSIR_OP_ITOI,
-    VSIR_OP_LABEL,
-    VSIR_OP_LD,
-    VSIR_OP_LD2DMS,
-    VSIR_OP_LD2DMS_S,
-    VSIR_OP_LD_RAW,
-    VSIR_OP_LD_RAW_S,
-    VSIR_OP_LD_S,
-    VSIR_OP_LD_STRUCTURED,
-    VSIR_OP_LD_STRUCTURED_S,
-    VSIR_OP_LD_UAV_TYPED,
-    VSIR_OP_LD_UAV_TYPED_S,
-    VSIR_OP_LIT,
-    VSIR_OP_LOD,
-    VSIR_OP_LOG,
-    VSIR_OP_LOGP,
-    VSIR_OP_LOOP,
-    VSIR_OP_LRP,
-    VSIR_OP_LTO,
-    VSIR_OP_LTU,
-    VSIR_OP_M3x2,
-    VSIR_OP_M3x3,
-    VSIR_OP_M3x4,
-    VSIR_OP_M4x3,
-    VSIR_OP_M4x4,
-    VSIR_OP_MAD,
-    VSIR_OP_MAX,
-    VSIR_OP_MIN,
-    VSIR_OP_MOV,
-    VSIR_OP_MOVA,
-    VSIR_OP_MOVC,
-    VSIR_OP_MSAD,
-    VSIR_OP_MUL,
-    VSIR_OP_NEG,
-    VSIR_OP_NEO,
-    VSIR_OP_NEU,
-    VSIR_OP_NOP,
-    VSIR_OP_NOT,
-    VSIR_OP_NRM,
-    VSIR_OP_OR,
-    VSIR_OP_ORD,
-    VSIR_OP_PHASE,
-    VSIR_OP_PHI,
-    VSIR_OP_POW,
-    VSIR_OP_QUAD_READ_ACROSS_D,
-    VSIR_OP_QUAD_READ_ACROSS_X,
-    VSIR_OP_QUAD_READ_ACROSS_Y,
-    VSIR_OP_QUAD_READ_LANE_AT,
-    VSIR_OP_RCP,
-    VSIR_OP_REP,
-    VSIR_OP_RESINFO,
-    VSIR_OP_RET,
-    VSIR_OP_RETP,
-    VSIR_OP_ROUND_NE,
-    VSIR_OP_ROUND_NI,
-    VSIR_OP_ROUND_PI,
-    VSIR_OP_ROUND_Z,
-    VSIR_OP_RSQ,
-    VSIR_OP_SAMPLE,
-    VSIR_OP_SAMPLE_B,
-    VSIR_OP_SAMPLE_B_CL_S,
-    VSIR_OP_SAMPLE_C,
-    VSIR_OP_SAMPLE_C_CL_S,
-    VSIR_OP_SAMPLE_C_LZ,
-    VSIR_OP_SAMPLE_C_LZ_S,
-    VSIR_OP_SAMPLE_CL_S,
-    VSIR_OP_SAMPLE_GRAD,
-    VSIR_OP_SAMPLE_GRAD_CL_S,
-    VSIR_OP_SAMPLE_INFO,
-    VSIR_OP_SAMPLE_LOD,
-    VSIR_OP_SAMPLE_LOD_S,
-    VSIR_OP_SAMPLE_POS,
-    VSIR_OP_SATURATE,
-    VSIR_OP_SETP,
-    VSIR_OP_SGE,
-    VSIR_OP_SGN,
-    VSIR_OP_SIN,
-    VSIR_OP_SINCOS,
-    VSIR_OP_SLT,
-    VSIR_OP_SQRT,
-    VSIR_OP_STORE_RAW,
-    VSIR_OP_STORE_STRUCTURED,
-    VSIR_OP_STORE_UAV_TYPED,
-    VSIR_OP_SUB,
-    VSIR_OP_SWAPC,
-    VSIR_OP_SWITCH,
-    VSIR_OP_SWITCH_MONOLITHIC,
-    VSIR_OP_SYNC,
-    VSIR_OP_TAN,
-    VSIR_OP_TEX,
-    VSIR_OP_TEXBEM,
-    VSIR_OP_TEXBEML,
-    VSIR_OP_TEXCOORD,
-    VSIR_OP_TEXCRD,
-    VSIR_OP_TEXDEPTH,
-    VSIR_OP_TEXDP3,
-    VSIR_OP_TEXDP3TEX,
-    VSIR_OP_TEXKILL,
-    VSIR_OP_TEXLD,
-    VSIR_OP_TEXLDD,
-    VSIR_OP_TEXLDL,
-    VSIR_OP_TEXM3x2DEPTH,
-    VSIR_OP_TEXM3x2PAD,
-    VSIR_OP_TEXM3x2TEX,
-    VSIR_OP_TEXM3x3,
-    VSIR_OP_TEXM3x3DIFF,
-    VSIR_OP_TEXM3x3PAD,
-    VSIR_OP_TEXM3x3SPEC,
-    VSIR_OP_TEXM3x3TEX,
-    VSIR_OP_TEXM3x3VSPEC,
-    VSIR_OP_TEXREG2AR,
-    VSIR_OP_TEXREG2GB,
-    VSIR_OP_TEXREG2RGB,
-    VSIR_OP_UBFE,
-    VSIR_OP_UDIV,
-    VSIR_OP_UDIV_SIMPLE,
-    VSIR_OP_UGE,
-    VSIR_OP_ULT,
-    VSIR_OP_UMAX,
-    VSIR_OP_UMIN,
-    VSIR_OP_UMUL,
-    VSIR_OP_UNO,
-    VSIR_OP_UREM,
-    VSIR_OP_USHR,
-    VSIR_OP_UTOD,
-    VSIR_OP_UTOF,
-    VSIR_OP_UTOU,
-    VSIR_OP_WAVE_ACTIVE_ALL_EQUAL,
-    VSIR_OP_WAVE_ACTIVE_BALLOT,
-    VSIR_OP_WAVE_ACTIVE_BIT_AND,
-    VSIR_OP_WAVE_ACTIVE_BIT_OR,
-    VSIR_OP_WAVE_ACTIVE_BIT_XOR,
-    VSIR_OP_WAVE_ALL_BIT_COUNT,
-    VSIR_OP_WAVE_ALL_TRUE,
-    VSIR_OP_WAVE_ANY_TRUE,
-    VSIR_OP_WAVE_IS_FIRST_LANE,
-    VSIR_OP_WAVE_OP_ADD,
-    VSIR_OP_WAVE_OP_IMAX,
-    VSIR_OP_WAVE_OP_IMIN,
-    VSIR_OP_WAVE_OP_MAX,
-    VSIR_OP_WAVE_OP_MIN,
-    VSIR_OP_WAVE_OP_MUL,
-    VSIR_OP_WAVE_OP_UMAX,
-    VSIR_OP_WAVE_OP_UMIN,
-    VSIR_OP_WAVE_PREFIX_BIT_COUNT,
-    VSIR_OP_WAVE_READ_LANE_AT,
-    VSIR_OP_WAVE_READ_LANE_FIRST,
-    VSIR_OP_XOR,
-
-    VSIR_OP_INVALID,
-
-    VSIR_OP_COUNT,
-};
-
-const char *vsir_opcode_get_name(enum vkd3d_shader_opcode op, const char *error);
+    VKD3DSIH_ABS,
+    VKD3DSIH_ACOS,
+    VKD3DSIH_ADD,
+    VKD3DSIH_AND,
+    VKD3DSIH_ASIN,
+    VKD3DSIH_ATAN,
+    VKD3DSIH_ATOMIC_AND,
+    VKD3DSIH_ATOMIC_CMP_STORE,
+    VKD3DSIH_ATOMIC_IADD,
+    VKD3DSIH_ATOMIC_IMAX,
+    VKD3DSIH_ATOMIC_IMIN,
+    VKD3DSIH_ATOMIC_OR,
+    VKD3DSIH_ATOMIC_UMAX,
+    VKD3DSIH_ATOMIC_UMIN,
+    VKD3DSIH_ATOMIC_XOR,
+    VKD3DSIH_BEM,
+    VKD3DSIH_BFI,
+    VKD3DSIH_BFREV,
+    VKD3DSIH_BRANCH,
+    VKD3DSIH_BREAK,
+    VKD3DSIH_BREAKC,
+    VKD3DSIH_BREAKP,
+    VKD3DSIH_BUFINFO,
+    VKD3DSIH_CALL,
+    VKD3DSIH_CALLNZ,
+    VKD3DSIH_CASE,
+    VKD3DSIH_CHECK_ACCESS_FULLY_MAPPED,
+    VKD3DSIH_CMP,
+    VKD3DSIH_CND,
+    VKD3DSIH_CONTINUE,
+    VKD3DSIH_CONTINUEP,
+    VKD3DSIH_COUNTBITS,
+    VKD3DSIH_CRS,
+    VKD3DSIH_CUT,
+    VKD3DSIH_CUT_STREAM,
+    VKD3DSIH_DADD,
+    VKD3DSIH_DCL,
+    VKD3DSIH_DCL_CONSTANT_BUFFER,
+    VKD3DSIH_DCL_FUNCTION_BODY,
+    VKD3DSIH_DCL_FUNCTION_TABLE,
+    VKD3DSIH_DCL_GLOBAL_FLAGS,
+    VKD3DSIH_DCL_GS_INSTANCES,
+    VKD3DSIH_DCL_HS_FORK_PHASE_INSTANCE_COUNT,
+    VKD3DSIH_DCL_HS_JOIN_PHASE_INSTANCE_COUNT,
+    VKD3DSIH_DCL_HS_MAX_TESSFACTOR,
+    VKD3DSIH_DCL_IMMEDIATE_CONSTANT_BUFFER,
+    VKD3DSIH_DCL_INDEX_RANGE,
+    VKD3DSIH_DCL_INDEXABLE_TEMP,
+    VKD3DSIH_DCL_INPUT,
+    VKD3DSIH_DCL_INPUT_CONTROL_POINT_COUNT,
+    VKD3DSIH_DCL_INPUT_PRIMITIVE,
+    VKD3DSIH_DCL_INPUT_PS,
+    VKD3DSIH_DCL_INPUT_PS_SGV,
+    VKD3DSIH_DCL_INPUT_PS_SIV,
+    VKD3DSIH_DCL_INPUT_SGV,
+    VKD3DSIH_DCL_INPUT_SIV,
+    VKD3DSIH_DCL_INTERFACE,
+    VKD3DSIH_DCL_OUTPUT,
+    VKD3DSIH_DCL_OUTPUT_CONTROL_POINT_COUNT,
+    VKD3DSIH_DCL_OUTPUT_SIV,
+    VKD3DSIH_DCL_OUTPUT_TOPOLOGY,
+    VKD3DSIH_DCL_RESOURCE_RAW,
+    VKD3DSIH_DCL_RESOURCE_STRUCTURED,
+    VKD3DSIH_DCL_SAMPLER,
+    VKD3DSIH_DCL_STREAM,
+    VKD3DSIH_DCL_TEMPS,
+    VKD3DSIH_DCL_TESSELLATOR_DOMAIN,
+    VKD3DSIH_DCL_TESSELLATOR_OUTPUT_PRIMITIVE,
+    VKD3DSIH_DCL_TESSELLATOR_PARTITIONING,
+    VKD3DSIH_DCL_TGSM_RAW,
+    VKD3DSIH_DCL_TGSM_STRUCTURED,
+    VKD3DSIH_DCL_THREAD_GROUP,
+    VKD3DSIH_DCL_UAV_RAW,
+    VKD3DSIH_DCL_UAV_STRUCTURED,
+    VKD3DSIH_DCL_UAV_TYPED,
+    VKD3DSIH_DCL_VERTICES_OUT,
+    VKD3DSIH_DDIV,
+    VKD3DSIH_DEF,
+    VKD3DSIH_DEFAULT,
+    VKD3DSIH_DEFB,
+    VKD3DSIH_DEFI,
+    VKD3DSIH_DEQO,
+    VKD3DSIH_DFMA,
+    VKD3DSIH_DGEO,
+    VKD3DSIH_DISCARD,
+    VKD3DSIH_DIV,
+    VKD3DSIH_DLT,
+    VKD3DSIH_DMAX,
+    VKD3DSIH_DMIN,
+    VKD3DSIH_DMOV,
+    VKD3DSIH_DMOVC,
+    VKD3DSIH_DMUL,
+    VKD3DSIH_DNE,
+    VKD3DSIH_DP2,
+    VKD3DSIH_DP2ADD,
+    VKD3DSIH_DP3,
+    VKD3DSIH_DP4,
+    VKD3DSIH_DRCP,
+    VKD3DSIH_DST,
+    VKD3DSIH_DSX,
+    VKD3DSIH_DSX_COARSE,
+    VKD3DSIH_DSX_FINE,
+    VKD3DSIH_DSY,
+    VKD3DSIH_DSY_COARSE,
+    VKD3DSIH_DSY_FINE,
+    VKD3DSIH_DTOF,
+    VKD3DSIH_DTOI,
+    VKD3DSIH_DTOU,
+    VKD3DSIH_ELSE,
+    VKD3DSIH_EMIT,
+    VKD3DSIH_EMIT_STREAM,
+    VKD3DSIH_ENDIF,
+    VKD3DSIH_ENDLOOP,
+    VKD3DSIH_ENDREP,
+    VKD3DSIH_ENDSWITCH,
+    VKD3DSIH_EQO,
+    VKD3DSIH_EQU,
+    VKD3DSIH_EVAL_CENTROID,
+    VKD3DSIH_EVAL_SAMPLE_INDEX,
+    VKD3DSIH_EXP,
+    VKD3DSIH_EXPP,
+    VKD3DSIH_F16TOF32,
+    VKD3DSIH_F32TOF16,
+    VKD3DSIH_FCALL,
+    VKD3DSIH_FIRSTBIT_HI,
+    VKD3DSIH_FIRSTBIT_LO,
+    VKD3DSIH_FIRSTBIT_SHI,
+    VKD3DSIH_FRC,
+    VKD3DSIH_FREM,
+    VKD3DSIH_FTOD,
+    VKD3DSIH_FTOI,
+    VKD3DSIH_FTOU,
+    VKD3DSIH_GATHER4,
+    VKD3DSIH_GATHER4_C,
+    VKD3DSIH_GATHER4_C_S,
+    VKD3DSIH_GATHER4_PO,
+    VKD3DSIH_GATHER4_PO_C,
+    VKD3DSIH_GATHER4_PO_C_S,
+    VKD3DSIH_GATHER4_PO_S,
+    VKD3DSIH_GATHER4_S,
+    VKD3DSIH_GEO,
+    VKD3DSIH_GEU,
+    VKD3DSIH_HCOS,
+    VKD3DSIH_HS_CONTROL_POINT_PHASE,
+    VKD3DSIH_HS_DECLS,
+    VKD3DSIH_HS_FORK_PHASE,
+    VKD3DSIH_HS_JOIN_PHASE,
+    VKD3DSIH_HSIN,
+    VKD3DSIH_HTAN,
+    VKD3DSIH_IADD,
+    VKD3DSIH_IBFE,
+    VKD3DSIH_IDIV,
+    VKD3DSIH_IEQ,
+    VKD3DSIH_IF,
+    VKD3DSIH_IFC,
+    VKD3DSIH_IGE,
+    VKD3DSIH_ILT,
+    VKD3DSIH_IMAD,
+    VKD3DSIH_IMAX,
+    VKD3DSIH_IMIN,
+    VKD3DSIH_IMM_ATOMIC_ALLOC,
+    VKD3DSIH_IMM_ATOMIC_AND,
+    VKD3DSIH_IMM_ATOMIC_CMP_EXCH,
+    VKD3DSIH_IMM_ATOMIC_CONSUME,
+    VKD3DSIH_IMM_ATOMIC_EXCH,
+    VKD3DSIH_IMM_ATOMIC_IADD,
+    VKD3DSIH_IMM_ATOMIC_IMAX,
+    VKD3DSIH_IMM_ATOMIC_IMIN,
+    VKD3DSIH_IMM_ATOMIC_OR,
+    VKD3DSIH_IMM_ATOMIC_UMAX,
+    VKD3DSIH_IMM_ATOMIC_UMIN,
+    VKD3DSIH_IMM_ATOMIC_XOR,
+    VKD3DSIH_IMUL,
+    VKD3DSIH_INE,
+    VKD3DSIH_INEG,
+    VKD3DSIH_ISFINITE,
+    VKD3DSIH_ISHL,
+    VKD3DSIH_ISHR,
+    VKD3DSIH_ISINF,
+    VKD3DSIH_ISNAN,
+    VKD3DSIH_ITOD,
+    VKD3DSIH_ITOF,
+    VKD3DSIH_ITOI,
+    VKD3DSIH_LABEL,
+    VKD3DSIH_LD,
+    VKD3DSIH_LD2DMS,
+    VKD3DSIH_LD2DMS_S,
+    VKD3DSIH_LD_RAW,
+    VKD3DSIH_LD_RAW_S,
+    VKD3DSIH_LD_S,
+    VKD3DSIH_LD_STRUCTURED,
+    VKD3DSIH_LD_STRUCTURED_S,
+    VKD3DSIH_LD_UAV_TYPED,
+    VKD3DSIH_LD_UAV_TYPED_S,
+    VKD3DSIH_LIT,
+    VKD3DSIH_LOD,
+    VKD3DSIH_LOG,
+    VKD3DSIH_LOGP,
+    VKD3DSIH_LOOP,
+    VKD3DSIH_LRP,
+    VKD3DSIH_LTO,
+    VKD3DSIH_LTU,
+    VKD3DSIH_M3x2,
+    VKD3DSIH_M3x3,
+    VKD3DSIH_M3x4,
+    VKD3DSIH_M4x3,
+    VKD3DSIH_M4x4,
+    VKD3DSIH_MAD,
+    VKD3DSIH_MAX,
+    VKD3DSIH_MIN,
+    VKD3DSIH_MOV,
+    VKD3DSIH_MOVA,
+    VKD3DSIH_MOVC,
+    VKD3DSIH_MSAD,
+    VKD3DSIH_MUL,
+    VKD3DSIH_NEO,
+    VKD3DSIH_NEU,
+    VKD3DSIH_NOP,
+    VKD3DSIH_NOT,
+    VKD3DSIH_NRM,
+    VKD3DSIH_OR,
+    VKD3DSIH_ORD,
+    VKD3DSIH_PHASE,
+    VKD3DSIH_PHI,
+    VKD3DSIH_POW,
+    VKD3DSIH_QUAD_READ_ACROSS_D,
+    VKD3DSIH_QUAD_READ_ACROSS_X,
+    VKD3DSIH_QUAD_READ_ACROSS_Y,
+    VKD3DSIH_QUAD_READ_LANE_AT,
+    VKD3DSIH_RCP,
+    VKD3DSIH_REP,
+    VKD3DSIH_RESINFO,
+    VKD3DSIH_RET,
+    VKD3DSIH_RETP,
+    VKD3DSIH_ROUND_NE,
+    VKD3DSIH_ROUND_NI,
+    VKD3DSIH_ROUND_PI,
+    VKD3DSIH_ROUND_Z,
+    VKD3DSIH_RSQ,
+    VKD3DSIH_SAMPLE,
+    VKD3DSIH_SAMPLE_B,
+    VKD3DSIH_SAMPLE_B_CL_S,
+    VKD3DSIH_SAMPLE_C,
+    VKD3DSIH_SAMPLE_C_CL_S,
+    VKD3DSIH_SAMPLE_C_LZ,
+    VKD3DSIH_SAMPLE_C_LZ_S,
+    VKD3DSIH_SAMPLE_CL_S,
+    VKD3DSIH_SAMPLE_GRAD,
+    VKD3DSIH_SAMPLE_GRAD_CL_S,
+    VKD3DSIH_SAMPLE_INFO,
+    VKD3DSIH_SAMPLE_LOD,
+    VKD3DSIH_SAMPLE_LOD_S,
+    VKD3DSIH_SAMPLE_POS,
+    VKD3DSIH_SETP,
+    VKD3DSIH_SGE,
+    VKD3DSIH_SGN,
+    VKD3DSIH_SINCOS,
+    VKD3DSIH_SLT,
+    VKD3DSIH_SQRT,
+    VKD3DSIH_STORE_RAW,
+    VKD3DSIH_STORE_STRUCTURED,
+    VKD3DSIH_STORE_UAV_TYPED,
+    VKD3DSIH_SUB,
+    VKD3DSIH_SWAPC,
+    VKD3DSIH_SWITCH,
+    VKD3DSIH_SWITCH_MONOLITHIC,
+    VKD3DSIH_SYNC,
+    VKD3DSIH_TAN,
+    VKD3DSIH_TEX,
+    VKD3DSIH_TEXBEM,
+    VKD3DSIH_TEXBEML,
+    VKD3DSIH_TEXCOORD,
+    VKD3DSIH_TEXDEPTH,
+    VKD3DSIH_TEXDP3,
+    VKD3DSIH_TEXDP3TEX,
+    VKD3DSIH_TEXKILL,
+    VKD3DSIH_TEXLDD,
+    VKD3DSIH_TEXLDL,
+    VKD3DSIH_TEXM3x2DEPTH,
+    VKD3DSIH_TEXM3x2PAD,
+    VKD3DSIH_TEXM3x2TEX,
+    VKD3DSIH_TEXM3x3,
+    VKD3DSIH_TEXM3x3DIFF,
+    VKD3DSIH_TEXM3x3PAD,
+    VKD3DSIH_TEXM3x3SPEC,
+    VKD3DSIH_TEXM3x3TEX,
+    VKD3DSIH_TEXM3x3VSPEC,
+    VKD3DSIH_TEXREG2AR,
+    VKD3DSIH_TEXREG2GB,
+    VKD3DSIH_TEXREG2RGB,
+    VKD3DSIH_UBFE,
+    VKD3DSIH_UDIV,
+    VKD3DSIH_UGE,
+    VKD3DSIH_ULT,
+    VKD3DSIH_UMAX,
+    VKD3DSIH_UMIN,
+    VKD3DSIH_UMUL,
+    VKD3DSIH_UNO,
+    VKD3DSIH_USHR,
+    VKD3DSIH_UTOD,
+    VKD3DSIH_UTOF,
+    VKD3DSIH_UTOU,
+    VKD3DSIH_WAVE_ACTIVE_ALL_EQUAL,
+    VKD3DSIH_WAVE_ACTIVE_BALLOT,
+    VKD3DSIH_WAVE_ACTIVE_BIT_AND,
+    VKD3DSIH_WAVE_ACTIVE_BIT_OR,
+    VKD3DSIH_WAVE_ACTIVE_BIT_XOR,
+    VKD3DSIH_WAVE_ALL_BIT_COUNT,
+    VKD3DSIH_WAVE_ALL_TRUE,
+    VKD3DSIH_WAVE_ANY_TRUE,
+    VKD3DSIH_WAVE_IS_FIRST_LANE,
+    VKD3DSIH_WAVE_OP_ADD,
+    VKD3DSIH_WAVE_OP_IMAX,
+    VKD3DSIH_WAVE_OP_IMIN,
+    VKD3DSIH_WAVE_OP_MAX,
+    VKD3DSIH_WAVE_OP_MIN,
+    VKD3DSIH_WAVE_OP_MUL,
+    VKD3DSIH_WAVE_OP_UMAX,
+    VKD3DSIH_WAVE_OP_UMIN,
+    VKD3DSIH_WAVE_PREFIX_BIT_COUNT,
+    VKD3DSIH_WAVE_READ_LANE_AT,
+    VKD3DSIH_WAVE_READ_LANE_FIRST,
+    VKD3DSIH_XOR,
+
+    VKD3DSIH_INVALID,
+
+    VKD3DSIH_COUNT,
+};
 
 enum vkd3d_shader_register_type
 {
@@ -713,62 +692,46 @@ enum vkd3d_shader_register_precision
     VKD3D_SHADER_REGISTER_PRECISION_INVALID = ~0u,
 };
 
-enum vsir_data_type
+enum vkd3d_data_type
 {
-    VSIR_DATA_BOOL,
-
-    VSIR_DATA_F16,
-    VSIR_DATA_F32,
-    VSIR_DATA_F64,
-
-    VSIR_DATA_I8,
-    VSIR_DATA_I16,
-    VSIR_DATA_I32,
-    VSIR_DATA_I64,
-
-    VSIR_DATA_U8,
-    VSIR_DATA_U16,
-    VSIR_DATA_U32,
-    VSIR_DATA_U64,
+    VKD3D_DATA_FLOAT,
+    VKD3D_DATA_INT,
+    VKD3D_DATA_UINT,
+    VKD3D_DATA_UNORM,
+    VKD3D_DATA_SNORM,
+    VKD3D_DATA_OPAQUE,
+    VKD3D_DATA_MIXED,
+    VKD3D_DATA_DOUBLE,
+    VKD3D_DATA_CONTINUED,
+    VKD3D_DATA_UNUSED,
+    VKD3D_DATA_UINT8,
+    VKD3D_DATA_UINT64,
+    VKD3D_DATA_BOOL,
+    VKD3D_DATA_UINT16,
+    VKD3D_DATA_HALF,
 
-    VSIR_DATA_SNORM,
-    VSIR_DATA_UNORM,
-
-    VSIR_DATA_OPAQUE,
-    VSIR_DATA_MIXED,
-    VSIR_DATA_CONTINUED,
-    VSIR_DATA_UNUSED,
-
-    VSIR_DATA_TYPE_COUNT,
+    VKD3D_DATA_COUNT,
 };
 
-const char *vsir_data_type_get_name(enum vsir_data_type t, const char *error);
-
-static inline bool data_type_is_integer(enum vsir_data_type data_type)
+static inline bool data_type_is_integer(enum vkd3d_data_type data_type)
 {
-    return data_type == VSIR_DATA_I8
-            || data_type == VSIR_DATA_I16
-            || data_type == VSIR_DATA_I32
-            || data_type == VSIR_DATA_I64
-            || data_type == VSIR_DATA_U8
-            || data_type == VSIR_DATA_U16
-            || data_type == VSIR_DATA_U32
-            || data_type == VSIR_DATA_U64;
+    return data_type == VKD3D_DATA_INT || data_type == VKD3D_DATA_UINT8 || data_type == VKD3D_DATA_UINT16
+            || data_type == VKD3D_DATA_UINT || data_type == VKD3D_DATA_UINT64;
 }
 
-static inline bool data_type_is_bool(enum vsir_data_type data_type)
+static inline bool data_type_is_bool(enum vkd3d_data_type data_type)
 {
-    return data_type == VSIR_DATA_BOOL;
+    return data_type == VKD3D_DATA_BOOL;
 }
 
-static inline bool data_type_is_floating_point(enum vsir_data_type data_type)
+static inline bool data_type_is_floating_point(enum vkd3d_data_type data_type)
 {
-    return data_type == VSIR_DATA_F16 || data_type == VSIR_DATA_F32 || data_type == VSIR_DATA_F64;
+    return data_type == VKD3D_DATA_HALF || data_type == VKD3D_DATA_FLOAT || data_type == VKD3D_DATA_DOUBLE;
 }
 
-static inline bool data_type_is_64_bit(enum vsir_data_type data_type)
+static inline bool data_type_is_64_bit(enum vkd3d_data_type data_type)
 {
-    return data_type == VSIR_DATA_F64 || data_type == VSIR_DATA_I64 || data_type == VSIR_DATA_U64;
+    return data_type == VKD3D_DATA_DOUBLE || data_type == VKD3D_DATA_UINT64;
 }
 
 enum vsir_dimension
@@ -882,7 +845,8 @@ enum vkd3d_shader_uav_flags
 
 enum vkd3d_shader_atomic_rmw_flags
 {
-    VKD3DARF_VOLATILE = 0x1,
+    VKD3DARF_SEQ_CST  = 0x1,
+    VKD3DARF_VOLATILE = 0x2,
 };
 
 enum vkd3d_tessellator_domain
@@ -952,16 +916,6 @@ enum vkd3d_shader_type
 
 struct vkd3d_shader_message_context;
 
-struct vkd3d_shader_source_list
-{
-    const char **sources;
-    size_t capacity, count;
-};
-
-bool vkd3d_shader_source_list_append(struct vkd3d_shader_source_list *l, const char *source);
-void vkd3d_shader_source_list_cleanup(struct vkd3d_shader_source_list *l);
-void vkd3d_shader_source_list_init(struct vkd3d_shader_source_list *l);
-
 struct vkd3d_shader_version
 {
     enum vkd3d_shader_type type;
@@ -972,7 +926,7 @@ struct vkd3d_shader_version
 struct vkd3d_shader_immediate_constant_buffer
 {
     unsigned int register_idx;
-    enum vsir_data_type data_type;
+    enum vkd3d_data_type data_type;
     /* total count is element_count * component_count */
     unsigned int element_count;
     unsigned int component_count;
@@ -985,7 +939,7 @@ struct vkd3d_shader_indexable_temp
     unsigned int register_idx;
     unsigned int register_size;
     unsigned int alignment;
-    enum vsir_data_type data_type;
+    enum vkd3d_data_type data_type;
     unsigned int component_count;
     bool has_function_scope;
     const struct vkd3d_shader_immediate_constant_buffer *initialiser;
@@ -1004,13 +958,13 @@ struct vkd3d_shader_register
     enum vkd3d_shader_register_type type;
     enum vkd3d_shader_register_precision precision;
     bool non_uniform;
-    enum vsir_data_type data_type;
+    enum vkd3d_data_type data_type;
     struct vkd3d_shader_register_index idx[3];
     unsigned int idx_count;
     enum vsir_dimension dimension;
     /* known address alignment for optimisation, or zero */
     unsigned int alignment;
-    union vsir_immediate_constant
+    union
     {
         uint32_t immconst_u32[VKD3D_VEC4_SIZE];
         float immconst_f32[VKD3D_VEC4_SIZE];
@@ -1021,7 +975,7 @@ struct vkd3d_shader_register
 };
 
 void vsir_register_init(struct vkd3d_shader_register *reg, enum vkd3d_shader_register_type reg_type,
-        enum vsir_data_type data_type, unsigned int idx_count);
+        enum vkd3d_data_type data_type, unsigned int idx_count);
 
 static inline bool vsir_register_is_descriptor(const struct vkd3d_shader_register *reg)
 {
@@ -1038,20 +992,6 @@ static inline bool vsir_register_is_descriptor(const struct vkd3d_shader_registe
     }
 }
 
-static inline enum vkd3d_shader_register_type vsir_register_type_from_sysval_input(
-        enum vkd3d_shader_sysval_semantic sysval)
-{
-    switch (sysval)
-    {
-        case VKD3D_SHADER_SV_PRIMITIVE_ID:
-            return VKD3DSPR_PRIMID;
-        case VKD3D_SHADER_SV_COVERAGE:
-            return VKD3DSPR_COVERAGE;
-        default:
-            return VKD3DSPR_INPUT;
-    }
-}
-
 struct vkd3d_shader_dst_param
 {
     struct vkd3d_shader_register reg;
@@ -1068,9 +1008,9 @@ struct vkd3d_shader_src_param
 };
 
 void vsir_src_param_init(struct vkd3d_shader_src_param *param, enum vkd3d_shader_register_type reg_type,
-        enum vsir_data_type data_type, unsigned int idx_count);
+        enum vkd3d_data_type data_type, unsigned int idx_count);
 void vsir_dst_param_init(struct vkd3d_shader_dst_param *param, enum vkd3d_shader_register_type reg_type,
-        enum vsir_data_type data_type, unsigned int idx_count);
+        enum vkd3d_data_type data_type, unsigned int idx_count);
 void vsir_dst_param_init_null(struct vkd3d_shader_dst_param *dst);
 void vsir_src_param_init_label(struct vkd3d_shader_src_param *param, unsigned int label_id);
 
@@ -1116,7 +1056,7 @@ struct vkd3d_shader_semantic
     unsigned int usage_idx;
     enum vkd3d_shader_resource_type resource_type;
     unsigned int sample_count;
-    enum vsir_data_type resource_data_type[VKD3D_VEC4_SIZE];
+    enum vkd3d_data_type resource_data_type[VKD3D_VEC4_SIZE];
     struct vkd3d_shader_resource resource;
 };
 
@@ -1322,14 +1262,14 @@ struct vkd3d_shader_instruction
     struct vkd3d_shader_location location;
     enum vkd3d_shader_opcode opcode;
     uint32_t flags;
-    size_t dst_count;
-    size_t src_count;
+    unsigned int dst_count;
+    unsigned int src_count;
     struct vkd3d_shader_dst_param *dst;
     struct vkd3d_shader_src_param *src;
     struct vkd3d_shader_texel_offset texel_offset;
     enum vkd3d_shader_resource_type resource_type;
     unsigned int resource_stride;
-    enum vsir_data_type resource_data_type[VKD3D_VEC4_SIZE];
+    enum vkd3d_data_type resource_data_type[VKD3D_VEC4_SIZE];
     bool coissue, structured, raw;
     const struct vkd3d_shader_src_param *predicate;
     union
@@ -1372,13 +1312,6 @@ static inline bool vkd3d_shader_ver_le(const struct vkd3d_shader_version *v, uns
 void vsir_instruction_init(struct vkd3d_shader_instruction *ins,
         const struct vkd3d_shader_location *location, enum vkd3d_shader_opcode opcode);
 
-static inline void vkd3d_shader_instruction_make_nop(struct vkd3d_shader_instruction *ins)
-{
-    struct vkd3d_shader_location location = ins->location;
-
-    vsir_instruction_init(ins, &location, VSIR_OP_NOP);
-}
-
 static inline bool vkd3d_shader_instruction_has_texel_offset(const struct vkd3d_shader_instruction *ins)
 {
     return ins->texel_offset.u || ins->texel_offset.v || ins->texel_offset.w;
@@ -1399,6 +1332,12 @@ static inline bool register_is_constant_or_undef(const struct vkd3d_shader_regis
     return register_is_constant(reg) || register_is_undef(reg);
 }
 
+static inline bool register_is_scalar_constant_zero(const struct vkd3d_shader_register *reg)
+{
+    return register_is_constant(reg) && reg->dimension == VSIR_DIMENSION_SCALAR
+            && (data_type_is_64_bit(reg->data_type) ? !reg->u.immconst_u64[0] : !reg->u.immconst_u32[0]);
+}
+
 static inline bool register_is_numeric_array(const struct vkd3d_shader_register *reg)
 {
     return (reg->type == VKD3DSPR_IMMCONSTBUFFER || reg->type == VKD3DSPR_IDXTEMP
@@ -1425,116 +1364,51 @@ struct vkd3d_shader_param_allocator
 {
     struct vkd3d_shader_param_node *head;
     struct vkd3d_shader_param_node *current;
-    size_t count;
-    size_t stride;
-    size_t index;
-};
-
-void *shader_param_allocator_get(struct vkd3d_shader_param_allocator *allocator, size_t count);
-
-struct vkd3d_shader_instruction_array
-{
-    struct vkd3d_shader_instruction *elements;
-    size_t capacity;
-    size_t count;
-};
-
-struct vkd3d_shader_instruction *shader_instruction_array_append(struct vkd3d_shader_instruction_array *array);
-bool shader_instruction_array_insert_at(struct vkd3d_shader_instruction_array *instructions, size_t idx, size_t count);
-
-struct vsir_program_iterator
-{
-    struct vkd3d_shader_instruction_array *array;
-    size_t idx;
+    unsigned int count;
+    unsigned int stride;
+    unsigned int index;
 };
 
-static inline struct vsir_program_iterator vsir_program_iterator(struct vkd3d_shader_instruction_array *array)
-{
-    return (struct vsir_program_iterator){ .array = array };
-}
-
-static inline struct vkd3d_shader_instruction *vsir_program_iterator_current(
-        struct vsir_program_iterator *iterator)
-{
-    if (iterator->idx >= iterator->array->count)
-        return NULL;
-
-    return &iterator->array->elements[iterator->idx];
-}
+void *shader_param_allocator_get(struct vkd3d_shader_param_allocator *allocator, unsigned int count);
 
-static inline struct vkd3d_shader_instruction *vsir_program_iterator_head(
-        struct vsir_program_iterator *iterator)
+static inline struct vkd3d_shader_src_param *shader_src_param_allocator_get(
+        struct vkd3d_shader_param_allocator *allocator, unsigned int count)
 {
-    iterator->idx = 0;
-
-    return vsir_program_iterator_current(iterator);
-}
-
-static inline struct vkd3d_shader_instruction *vsir_program_iterator_tail(struct vsir_program_iterator *iterator)
-{
-    iterator->idx = iterator->array->count - 1;
-
-    return vsir_program_iterator_current(iterator);
-}
-
-static inline struct vkd3d_shader_instruction *vsir_program_iterator_next(
-        struct vsir_program_iterator *iterator)
-{
-    if (iterator->idx < iterator->array->count || iterator->idx == SIZE_MAX)
-        ++iterator->idx;
-
-    return vsir_program_iterator_current(iterator);
-}
-
-static inline struct vkd3d_shader_instruction *vsir_program_iterator_prev(
-        struct vsir_program_iterator *iterator)
-{
-    if (iterator->idx != SIZE_MAX)
-        --iterator->idx;
-
-    return vsir_program_iterator_current(iterator);
+    VKD3D_ASSERT(allocator->stride == sizeof(struct vkd3d_shader_src_param));
+    return shader_param_allocator_get(allocator, count);
 }
 
-/* When insertion takes place, argument `it' is updated to point to the same
- * instruction as before the insertion, but all existing pointers to the same
- * container, as well as any iterators pointing to instructions after the
- * insertion point should be considered invalid. */
-static inline bool vsir_program_iterator_insert_after(struct vsir_program_iterator *it, size_t count)
+static inline struct vkd3d_shader_dst_param *shader_dst_param_allocator_get(
+        struct vkd3d_shader_param_allocator *allocator, unsigned int count)
 {
-    return shader_instruction_array_insert_at(it->array, it->idx + 1, count);
+    VKD3D_ASSERT(allocator->stride == sizeof(struct vkd3d_shader_dst_param));
+    return shader_param_allocator_get(allocator, count);
 }
 
-/* When insertion takes place, argument `it' is updated to point to the same
- * instruction as before the insertion, and argument `ins_it' is initialized
- * to point to the first inserted instruction. A pointer to the first inserted
- * instruction is returned. */
-static inline struct vkd3d_shader_instruction *vsir_program_iterator_insert_before(
-        struct vsir_program_iterator *it, struct vsir_program_iterator *ins_it, size_t count)
+struct vkd3d_shader_instruction_array
 {
-    VKD3D_ASSERT(it != ins_it);
-    VKD3D_ASSERT(it->idx != SIZE_MAX);
-
-    if (!shader_instruction_array_insert_at(it->array, it->idx, count))
-        return NULL;
-
-    *ins_it = *it;
-    it->idx += count;
+    struct vkd3d_shader_instruction *elements;
+    size_t capacity;
+    size_t count;
 
-    return vsir_program_iterator_current(ins_it);
-}
+    struct vkd3d_shader_param_allocator src_params;
+    struct vkd3d_shader_param_allocator dst_params;
+    struct vkd3d_shader_immediate_constant_buffer **icbs;
+    size_t icb_capacity;
+    size_t icb_count;
 
-/* When insertion takes place, argument `it' is updated to point to the first
- * inserted instruction. A pointer to this first inserted instruction is
- * returned. */
-static inline struct vkd3d_shader_instruction *vsir_program_iterator_insert_before_and_move(
-        struct vsir_program_iterator *it, size_t count)
-{
-    VKD3D_ASSERT(it->idx != SIZE_MAX);
+    struct vkd3d_shader_src_param *outpointid_param;
+};
 
-    if (!shader_instruction_array_insert_at(it->array, it->idx, count))
-        return NULL;
-    return vsir_program_iterator_current(it);
-}
+bool shader_instruction_array_init(struct vkd3d_shader_instruction_array *instructions, unsigned int reserve);
+bool shader_instruction_array_reserve(struct vkd3d_shader_instruction_array *instructions, unsigned int reserve);
+bool shader_instruction_array_insert_at(struct vkd3d_shader_instruction_array *instructions,
+        unsigned int idx, unsigned int count);
+bool shader_instruction_array_add_icb(struct vkd3d_shader_instruction_array *instructions,
+        struct vkd3d_shader_immediate_constant_buffer *icb);
+bool shader_instruction_array_clone_instruction(struct vkd3d_shader_instruction_array *instructions,
+        unsigned int dst, unsigned int src);
+void shader_instruction_array_destroy(struct vkd3d_shader_instruction_array *instructions);
 
 enum vkd3d_shader_config_flags
 {
@@ -1562,7 +1436,7 @@ struct vkd3d_shader_descriptor_info1
     unsigned int register_index;
     unsigned int register_id;
     enum vkd3d_shader_resource_type resource_type;
-    enum vsir_data_type resource_data_type;
+    enum vkd3d_data_type resource_data_type;
     unsigned int flags;
     unsigned int sample_count;
     unsigned int buffer_size;
@@ -1593,7 +1467,6 @@ struct vsir_program
 
     struct vkd3d_shader_scan_descriptor_info1 descriptors;
     bool has_descriptor_info;
-    size_t descriptors_size;
 
     unsigned int parameter_count;
     const struct vkd3d_shader_parameter1 *parameters;
@@ -1601,7 +1474,8 @@ struct vsir_program
 
     unsigned int input_control_point_count, output_control_point_count;
     struct vsir_thread_group_size thread_group_size;
-    unsigned int block_count; /* maximum block count in any function */
+    unsigned int flat_constant_count[3];
+    unsigned int block_count;
     unsigned int temp_count;
     unsigned int ssa_count;
     enum vsir_global_flags global_flags;
@@ -1612,7 +1486,6 @@ struct vsir_program
     uint8_t diffuse_written_mask;
     enum vsir_control_flow_type cf_type;
     enum vsir_normalisation_level normalisation_level;
-    bool has_no_modifiers;
     enum vkd3d_tessellator_domain tess_domain;
     enum vkd3d_shader_tessellator_partitioning tess_partitioning;
     enum vkd3d_shader_tessellator_output_primitive tess_output_primitive;
@@ -1623,38 +1496,19 @@ struct vsir_program
 
     struct vsir_features features;
 
-    struct vkd3d_shader_source_list source_files;
     const char **block_names;
     size_t block_name_count;
-
-    struct vkd3d_shader_immediate_constant_buffer **icbs;
-    size_t icb_capacity;
-    size_t icb_count;
-
-    struct vkd3d_shader_param_allocator src_params;
-    struct vkd3d_shader_param_allocator dst_params;
 };
 
-enum vkd3d_result vsir_allocate_temp_registers(struct vsir_program *program,
-        struct vkd3d_shader_message_context *message_context);
-enum vkd3d_result vsir_update_dcl_temps(struct vsir_program *program,
-        struct vkd3d_shader_message_context *message_context);
-
-struct vkd3d_shader_descriptor_info1 *vsir_program_add_descriptor(struct vsir_program *program,
-        enum vkd3d_shader_descriptor_type type, unsigned int register_id,
-        const struct vkd3d_shader_register_range *range,
-        enum vkd3d_shader_resource_type resource_type, enum vsir_data_type resource_data_type);
-bool vsir_program_add_icb(struct vsir_program *program, struct vkd3d_shader_immediate_constant_buffer *icb);
 void vsir_program_cleanup(struct vsir_program *program);
+int vsir_program_compile(struct vsir_program *program, uint64_t config_flags,
+        const struct vkd3d_shader_compile_info *compile_info, struct vkd3d_shader_code *out,
+        struct vkd3d_shader_message_context *message_context);
 const struct vkd3d_shader_parameter1 *vsir_program_get_parameter(
         const struct vsir_program *program, enum vkd3d_shader_parameter_name name);
 bool vsir_program_init(struct vsir_program *program, const struct vkd3d_shader_compile_info *compile_info,
         const struct vkd3d_shader_version *version, unsigned int reserve, enum vsir_control_flow_type cf_type,
         enum vsir_normalisation_level normalisation_level);
-enum vkd3d_result vsir_program_lower_d3dbc(struct vsir_program *program, uint64_t config_flags,
-        const struct vkd3d_shader_compile_info *compile_info, struct vkd3d_shader_message_context *message_context);
-enum vkd3d_result vsir_program_optimize(struct vsir_program *program, uint64_t config_flags,
-        const struct vkd3d_shader_compile_info *compile_info, struct vkd3d_shader_message_context *message_context);
 enum vkd3d_result vsir_program_transform(struct vsir_program *program, uint64_t config_flags,
         const struct vkd3d_shader_compile_info *compile_info, struct vkd3d_shader_message_context *message_context);
 enum vkd3d_result vsir_program_transform_early(struct vsir_program *program, uint64_t config_flags,
@@ -1667,46 +1521,34 @@ bool vsir_instruction_init_with_params(struct vsir_program *program,
         struct vkd3d_shader_instruction *ins, const struct vkd3d_shader_location *location,
         enum vkd3d_shader_opcode opcode, unsigned int dst_count, unsigned int src_count);
 
-static inline struct vkd3d_shader_instruction *vsir_program_append(struct vsir_program *program)
-{
-    return shader_instruction_array_append(&program->instructions);
-}
-
 static inline struct vkd3d_shader_dst_param *vsir_program_get_dst_params(
         struct vsir_program *program, unsigned int count)
 {
-    struct vkd3d_shader_param_allocator *allocator = &program->dst_params;
-
-    VKD3D_ASSERT(allocator->stride == sizeof(struct vkd3d_shader_dst_param));
-
-    return shader_param_allocator_get(allocator, count);
+    return shader_dst_param_allocator_get(&program->instructions.dst_params, count);
 }
 
 static inline struct vkd3d_shader_src_param *vsir_program_get_src_params(
         struct vsir_program *program, unsigned int count)
 {
-    struct vkd3d_shader_param_allocator *allocator = &program->src_params;
-
-    VKD3D_ASSERT(allocator->stride == sizeof(struct vkd3d_shader_src_param));
-
-    return shader_param_allocator_get(allocator, count);
+    return shader_src_param_allocator_get(&program->instructions.src_params, count);
 }
 
 struct vkd3d_shader_parser
 {
     struct vkd3d_shader_message_context *message_context;
     struct vkd3d_shader_location location;
-    enum vkd3d_result status;
+    struct vsir_program *program;
+    bool failed;
 };
 
 void vkd3d_shader_parser_error(struct vkd3d_shader_parser *parser,
         enum vkd3d_shader_error error, const char *format, ...) VKD3D_PRINTF_FUNC(3, 4);
-void vkd3d_shader_parser_init(struct vkd3d_shader_parser *parser,
+void vkd3d_shader_parser_init(struct vkd3d_shader_parser *parser, struct vsir_program *program,
         struct vkd3d_shader_message_context *message_context, const char *source_name);
 void vkd3d_shader_parser_warning(struct vkd3d_shader_parser *parser,
         enum vkd3d_shader_error error, const char *format, ...) VKD3D_PRINTF_FUNC(3, 4);
 
-void vsir_program_trace(struct vsir_program *program);
+void vsir_program_trace(const struct vsir_program *program);
 
 const char *shader_get_type_prefix(enum vkd3d_shader_type type);
 
@@ -1729,12 +1571,11 @@ enum vsir_asm_flags
     VSIR_ASM_FLAG_DUMP_ALL_INDICES = 0x2,
     VSIR_ASM_FLAG_DUMP_SIGNATURES = 0x4,
     VSIR_ASM_FLAG_DUMP_DESCRIPTORS = 0x8,
-    VSIR_ASM_FLAG_ALLOCATE_TEMPS = 0x10,
 };
 
-enum vkd3d_result d3d_asm_compile(struct vsir_program *program,
-        const struct vkd3d_shader_compile_info *compile_info, struct vkd3d_shader_code *out,
-        enum vsir_asm_flags flags, struct vkd3d_shader_message_context *message_context);
+enum vkd3d_result d3d_asm_compile(const struct vsir_program *program,
+        const struct vkd3d_shader_compile_info *compile_info,
+        struct vkd3d_shader_code *out, enum vsir_asm_flags flags);
 void vkd3d_string_buffer_cleanup(struct vkd3d_string_buffer *buffer);
 struct vkd3d_string_buffer *vkd3d_string_buffer_get(struct vkd3d_string_buffer_cache *list);
 void vkd3d_string_buffer_init(struct vkd3d_string_buffer *buffer);
@@ -1841,8 +1682,6 @@ int tpf_parse(const struct vkd3d_shader_compile_info *compile_info, uint64_t con
         struct vkd3d_shader_message_context *message_context, struct vsir_program *program);
 int fx_parse(const struct vkd3d_shader_compile_info *compile_info,
         struct vkd3d_shader_code *out, struct vkd3d_shader_message_context *message_context);
-int tx_parse(const struct vkd3d_shader_compile_info *compile_info,
-        struct vkd3d_shader_code *out, struct vkd3d_shader_message_context *message_context);
 
 void free_dxbc_shader_desc(struct dxbc_shader_desc *desc);
 
@@ -1870,8 +1709,7 @@ int msl_compile(struct vsir_program *program, uint64_t config_flags,
         const struct vkd3d_shader_compile_info *compile_info, struct vkd3d_shader_code *out,
         struct vkd3d_shader_message_context *message_context);
 
-int tpf_compile(struct vsir_program *program, uint64_t config_flags,
-        const struct vkd3d_shader_compile_info *compile_info, const struct vkd3d_shader_code *rdef,
+int tpf_compile(struct vsir_program *program, uint64_t config_flags, const struct vkd3d_shader_code *rdef,
         struct vkd3d_shader_code *out, struct vkd3d_shader_message_context *message_context);
 
 enum vkd3d_md5_variant
@@ -1885,45 +1723,68 @@ void vkd3d_compute_md5(const void *dxbc, size_t size, uint32_t checksum[4], enum
 int preproc_lexer_parse(const struct vkd3d_shader_compile_info *compile_info,
         struct vkd3d_shader_code *out, struct vkd3d_shader_message_context *message_context);
 
-int hlsl_compile_effect(const struct vkd3d_shader_compile_info *compile_info,
-        struct vkd3d_shader_message_context *message_context, struct vkd3d_shader_code *out);
-int hlsl_parse(const struct vkd3d_shader_compile_info *compile_info,
-        struct vkd3d_shader_message_context *message_context,
-        struct vsir_program *program, struct vkd3d_shader_code *reflection_data);
+int hlsl_compile_shader(const struct vkd3d_shader_code *hlsl, const struct vkd3d_shader_compile_info *compile_info,
+        struct vkd3d_shader_code *out, struct vkd3d_shader_message_context *message_context);
 
-static inline enum vsir_data_type vsir_data_type_from_component_type(enum vkd3d_shader_component_type component_type)
+static inline enum vkd3d_shader_component_type vkd3d_component_type_from_data_type(
+        enum vkd3d_data_type data_type)
+{
+    switch (data_type)
+    {
+        case VKD3D_DATA_HALF: /* Minimum precision. TODO: native 16-bit */
+        case VKD3D_DATA_FLOAT:
+        case VKD3D_DATA_UNORM:
+        case VKD3D_DATA_SNORM:
+            return VKD3D_SHADER_COMPONENT_FLOAT;
+        case VKD3D_DATA_UINT16: /* Minimum precision. TODO: native 16-bit */
+        case VKD3D_DATA_UINT:
+            return VKD3D_SHADER_COMPONENT_UINT;
+        case VKD3D_DATA_INT:
+            return VKD3D_SHADER_COMPONENT_INT;
+        case VKD3D_DATA_DOUBLE:
+            return VKD3D_SHADER_COMPONENT_DOUBLE;
+        case VKD3D_DATA_UINT64:
+            return VKD3D_SHADER_COMPONENT_UINT64;
+        case VKD3D_DATA_BOOL:
+            return VKD3D_SHADER_COMPONENT_BOOL;
+        default:
+            FIXME("Unhandled data type %#x.\n", data_type);
+            /* fall-through */
+        case VKD3D_DATA_MIXED:
+            return VKD3D_SHADER_COMPONENT_UINT;
+    }
+}
+
+static inline enum vkd3d_data_type vkd3d_data_type_from_component_type(
+        enum vkd3d_shader_component_type component_type)
 {
     switch (component_type)
     {
-        case VKD3D_SHADER_COMPONENT_VOID:
-            return VSIR_DATA_UNUSED;
+        case VKD3D_SHADER_COMPONENT_FLOAT:
+            return VKD3D_DATA_FLOAT;
         case VKD3D_SHADER_COMPONENT_UINT:
-            return VSIR_DATA_U32;
+            return VKD3D_DATA_UINT;
         case VKD3D_SHADER_COMPONENT_INT:
-            return VSIR_DATA_I32;
-        case VKD3D_SHADER_COMPONENT_FLOAT:
-            return VSIR_DATA_F32;
-        case VKD3D_SHADER_COMPONENT_BOOL:
-            return VSIR_DATA_BOOL;
+            return VKD3D_DATA_INT;
         case VKD3D_SHADER_COMPONENT_DOUBLE:
-            return VSIR_DATA_F64;
-        case VKD3D_SHADER_COMPONENT_UINT64:
-            return VSIR_DATA_U64;
-        case VKD3D_SHADER_COMPONENT_INT64:
-            return VSIR_DATA_I64;
-        case VKD3D_SHADER_COMPONENT_FLOAT16:
-            return VSIR_DATA_F16;
-        case VKD3D_SHADER_COMPONENT_UINT16:
-            return VSIR_DATA_U16;
-        case VKD3D_SHADER_COMPONENT_INT16:
-            return VSIR_DATA_I16;
-        case VKD3D_SHADER_COMPONENT_TYPE_FORCE_32BIT:
-            break;
+            return VKD3D_DATA_DOUBLE;
+        default:
+            FIXME("Unhandled component type %#x.\n", component_type);
+            return VKD3D_DATA_FLOAT;
     }
+}
 
-    FIXME("Unhandled component type %#x.\n", component_type);
+static inline bool component_type_is_64_bit(enum vkd3d_shader_component_type component_type)
+{
+    return component_type == VKD3D_SHADER_COMPONENT_DOUBLE || component_type == VKD3D_SHADER_COMPONENT_UINT64;
+}
+
+enum vkd3d_shader_input_sysval_semantic vkd3d_siv_from_sysval_indexed(enum vkd3d_shader_sysval_semantic sysval,
+        unsigned int index);
 
-    return VSIR_DATA_UNUSED;
+static inline enum vkd3d_shader_input_sysval_semantic vkd3d_siv_from_sysval(enum vkd3d_shader_sysval_semantic sysval)
+{
+    return vkd3d_siv_from_sysval_indexed(sysval, 0);
 }
 
 static inline unsigned int vsir_write_mask_get_component_idx(uint32_t write_mask)
@@ -1992,8 +1853,6 @@ static inline uint32_t vsir_write_mask_32_from_64(uint32_t write_mask64)
     }
 }
 
-const char *debug_vsir_writemask(unsigned int writemask);
-
 static inline uint32_t vsir_swizzle_64_from_32(uint32_t swizzle32)
 {
     switch (swizzle32)
@@ -2043,12 +1902,6 @@ static inline unsigned int vsir_swizzle_get_component(uint32_t swizzle, unsigned
     return (swizzle >> VKD3D_SHADER_SWIZZLE_SHIFT(idx)) & VKD3D_SHADER_SWIZZLE_MASK;
 }
 
-static inline void vsir_swizzle_set_component(uint32_t *swizzle, unsigned int idx, unsigned int component)
-{
-    *swizzle &= ~(VKD3D_SHADER_SWIZZLE_MASK << VKD3D_SHADER_SWIZZLE_SHIFT(idx));
-    *swizzle |= component << VKD3D_SHADER_SWIZZLE_SHIFT(idx);
-}
-
 static inline unsigned int vkd3d_compact_swizzle(uint32_t swizzle, uint32_t write_mask)
 {
     unsigned int i, compacted_swizzle = 0;
