diff --git a/libs/vkd3d/libs/vkd3d/command.c b/libs/vkd3d/libs/vkd3d/command.c
index 69f42280e8a749c2b266254e5c43a377a9d2bac1..e487ed0b9ad6a46e73384116d345688b96ff4772 100644
--- a/libs/vkd3d/libs/vkd3d/command.c
+++ b/libs/vkd3d/libs/vkd3d/command.c
@@ -2688,12 +2688,32 @@ static void d3d12_command_list_reset_state(struct d3d12_command_list *list,
 static HRESULT STDMETHODCALLTYPE d3d12_command_list_Reset(ID3D12GraphicsCommandList6 *iface,
         ID3D12CommandAllocator *allocator, ID3D12PipelineState *initial_pipeline_state)
 {
+    struct d3d12_command_allocator *allocator_impl = unsafe_impl_from_ID3D12CommandAllocator(allocator);
     struct d3d12_command_list *list = impl_from_ID3D12GraphicsCommandList6(iface);
+    HRESULT hr;
 
     TRACE("iface %p, allocator %p, initial_pipeline_state %p.\n",
             iface, allocator, initial_pipeline_state);
 
-    return d3d12_command_list_reset(list, allocator, initial_pipeline_state);
+    if (!allocator_impl)
+    {
+        WARN("Command allocator is NULL.\n");
+        return E_INVALIDARG;
+    }
+
+    if (list->is_recording)
+    {
+        WARN("Command list is in the recording state.\n");
+        return E_FAIL;
+    }
+
+    if (SUCCEEDED(hr = d3d12_command_allocator_allocate_command_buffer(allocator_impl, list)))
+    {
+        list->allocator = allocator_impl;
+        d3d12_command_list_reset_state(list, initial_pipeline_state);
+    }
+
+    return hr;
 }
 
 static void STDMETHODCALLTYPE d3d12_command_list_ClearState(ID3D12GraphicsCommandList6 *iface,
@@ -3029,7 +3049,7 @@ static void d3d12_command_list_update_descriptor_table(struct d3d12_command_list
 
             if (descriptor_count > range->vk_binding_count)
             {
-                MESSAGE("Heap descriptor count %u exceeds maximum Vulkan count %u. Reducing to the Vulkan maximum.\n",
+                ERR("Heap descriptor count %u exceeds maximum Vulkan count %u. Reducing to the Vulkan maximum.\n",
                         descriptor_count, range->vk_binding_count);
                 descriptor_count = range->vk_binding_count;
             }
@@ -4201,21 +4221,13 @@ static void STDMETHODCALLTYPE d3d12_command_list_RSSetViewports(ID3D12GraphicsCo
     TRACE("iface %p, viewport_count %u, viewports %p.\n", iface, viewport_count, viewports);
 
     if (viewport_count > ARRAY_SIZE(vk_viewports))
+    {
         FIXME("Viewport count %u > D3D12_VIEWPORT_AND_SCISSORRECT_OBJECT_COUNT_PER_PIPELINE.\n", viewport_count);
+        viewport_count = ARRAY_SIZE(vk_viewports);
+    }
 
-    for (i = 0; i < ARRAY_SIZE(vk_viewports); ++i)
+    for (i = 0; i < viewport_count; ++i)
     {
-        if (i >= viewport_count)
-        {
-            vk_viewports[i].x = 0.0f;
-            vk_viewports[i].y = 0.0f;
-            vk_viewports[i].width = 1.0f;
-            vk_viewports[i].height = 1.0f;
-            vk_viewports[i].minDepth = 0.0f;
-            vk_viewports[i].maxDepth = 0.0f;
-            continue;
-        }
-
         vk_viewports[i].x = viewports[i].TopLeftX;
         vk_viewports[i].y = viewports[i].TopLeftY + viewports[i].Height;
         vk_viewports[i].width = viewports[i].Width;
@@ -4233,7 +4245,7 @@ static void STDMETHODCALLTYPE d3d12_command_list_RSSetViewports(ID3D12GraphicsCo
     }
 
     vk_procs = &list->device->vk_procs;
-    VK_CALL(vkCmdSetViewport(list->vk_command_buffer, 0, ARRAY_SIZE(vk_viewports), vk_viewports));
+    VK_CALL(vkCmdSetViewport(list->vk_command_buffer, 0, viewport_count, vk_viewports));
 }
 
 static void STDMETHODCALLTYPE d3d12_command_list_RSSetScissorRects(ID3D12GraphicsCommandList6 *iface,
@@ -4252,7 +4264,6 @@ static void STDMETHODCALLTYPE d3d12_command_list_RSSetScissorRects(ID3D12Graphic
         rect_count = ARRAY_SIZE(vk_rects);
     }
 
-    memset(vk_rects, 0, sizeof(vk_rects));
     for (i = 0; i < rect_count; ++i)
     {
         vk_rects[i].offset.x = rects[i].left;
@@ -4262,7 +4273,7 @@ static void STDMETHODCALLTYPE d3d12_command_list_RSSetScissorRects(ID3D12Graphic
     }
 
     vk_procs = &list->device->vk_procs;
-    VK_CALL(vkCmdSetScissor(list->vk_command_buffer, 0, ARRAY_SIZE(vk_rects), vk_rects));
+    VK_CALL(vkCmdSetScissor(list->vk_command_buffer, 0, rect_count, vk_rects));
 }
 
 static void STDMETHODCALLTYPE d3d12_command_list_OMSetBlendFactor(ID3D12GraphicsCommandList6 *iface,
@@ -6385,8 +6396,9 @@ static struct d3d12_command_list *unsafe_impl_from_ID3D12CommandList(ID3D12Comma
     return CONTAINING_RECORD(iface, struct d3d12_command_list, ID3D12GraphicsCommandList6_iface);
 }
 
-static HRESULT d3d12_command_list_init(struct d3d12_command_list *list,
-        struct d3d12_device *device, D3D12_COMMAND_LIST_TYPE type)
+static HRESULT d3d12_command_list_init(struct d3d12_command_list *list, struct d3d12_device *device,
+        D3D12_COMMAND_LIST_TYPE type, struct d3d12_command_allocator *allocator,
+        ID3D12PipelineState *initial_pipeline_state)
 {
     HRESULT hr;
 
@@ -6400,37 +6412,31 @@ static HRESULT d3d12_command_list_init(struct d3d12_command_list *list,
 
     d3d12_device_add_ref(list->device = device);
 
-    return hr;
-}
-
-HRESULT d3d12_command_list_create(struct d3d12_device *device, UINT node_mask,
-        D3D12_COMMAND_LIST_TYPE type, struct d3d12_command_list **list)
-{
-    struct d3d12_command_list *object;
-    HRESULT hr;
+    list->allocator = allocator;
 
-    debug_ignored_node_mask(node_mask);
-
-    if (!(object = vkd3d_calloc(1, sizeof(*object))))
-        return E_OUTOFMEMORY;
+    list->descriptor_heap_count = 0;
 
-    if (FAILED(hr = d3d12_command_list_init(object, device, type)))
+    if (SUCCEEDED(hr = d3d12_command_allocator_allocate_command_buffer(allocator, list)))
     {
-        vkd3d_free(object);
-        return hr;
+        list->pipeline_bindings[VKD3D_PIPELINE_BIND_POINT_GRAPHICS].vk_uav_counter_views = NULL;
+        list->pipeline_bindings[VKD3D_PIPELINE_BIND_POINT_COMPUTE].vk_uav_counter_views = NULL;
+        d3d12_command_list_reset_state(list, initial_pipeline_state);
+    }
+    else
+    {
+        vkd3d_private_store_destroy(&list->private_store);
+        d3d12_device_release(device);
     }
 
-    TRACE("Created command list %p.\n", object);
-
-    *list = object;
-
-    return S_OK;
+    return hr;
 }
 
-HRESULT d3d12_command_list_reset(struct d3d12_command_list *list, ID3D12CommandAllocator *allocator_iface,
-        ID3D12PipelineState *initial_pipeline_state)
+HRESULT d3d12_command_list_create(struct d3d12_device *device,
+        UINT node_mask, D3D12_COMMAND_LIST_TYPE type, ID3D12CommandAllocator *allocator_iface,
+        ID3D12PipelineState *initial_pipeline_state, struct d3d12_command_list **list)
 {
     struct d3d12_command_allocator *allocator;
+    struct d3d12_command_list *object;
     HRESULT hr;
 
     if (!(allocator = unsafe_impl_from_ID3D12CommandAllocator(allocator_iface)))
@@ -6439,26 +6445,29 @@ HRESULT d3d12_command_list_reset(struct d3d12_command_list *list, ID3D12CommandA
         return E_INVALIDARG;
     }
 
-    if (allocator->type != list->type)
+    if (allocator->type != type)
     {
         WARN("Command list types do not match (allocator %#x, list %#x).\n",
-                allocator->type, list->type);
+                allocator->type, type);
         return E_INVALIDARG;
     }
 
-    if (list->is_recording)
-    {
-        WARN("Command list is in the recording state.\n");
-        return E_FAIL;
-    }
+    debug_ignored_node_mask(node_mask);
 
-    if (SUCCEEDED(hr = d3d12_command_allocator_allocate_command_buffer(allocator, list)))
+    if (!(object = vkd3d_malloc(sizeof(*object))))
+        return E_OUTOFMEMORY;
+
+    if (FAILED(hr = d3d12_command_list_init(object, device, type, allocator, initial_pipeline_state)))
     {
-        list->allocator = allocator;
-        d3d12_command_list_reset_state(list, initial_pipeline_state);
+        vkd3d_free(object);
+        return hr;
     }
 
-    return hr;
+    TRACE("Created command list %p.\n", object);
+
+    *list = object;
+
+    return S_OK;
 }
 
 /* ID3D12CommandQueue */
