diff --git a/libs/vkd3d/libs/vkd3d/device.c b/libs/vkd3d/libs/vkd3d/device.c
index 6af5e2a5c7d2108b25045e6465a7869c44be6267..b2636fd55856dad93a3bff2ae5bf5374b5d99880 100644
--- a/libs/vkd3d/libs/vkd3d/device.c
+++ b/libs/vkd3d/libs/vkd3d/device.c
@@ -3271,15 +3271,10 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_CreateCommandList(ID3D12Device9 *i
             iface, node_mask, type, command_allocator,
             initial_pipeline_state, debugstr_guid(riid), command_list);
 
-    if (FAILED(hr = d3d12_command_list_create(device, node_mask, type, &object)))
+    if (FAILED(hr = d3d12_command_list_create(device, node_mask, type, command_allocator,
+            initial_pipeline_state, &object)))
         return hr;
 
-    if (FAILED(hr = d3d12_command_list_reset(object, command_allocator, initial_pipeline_state)))
-    {
-        ID3D12GraphicsCommandList6_Release(&object->ID3D12GraphicsCommandList6_iface);
-        return hr;
-    }
-
     return return_interface(&object->ID3D12GraphicsCommandList6_iface,
             &IID_ID3D12GraphicsCommandList6, riid, command_list);
 }
@@ -5087,21 +5082,10 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_CreateCommandList1(ID3D12Device9 *
         UINT node_mask, D3D12_COMMAND_LIST_TYPE type, D3D12_COMMAND_LIST_FLAGS flags,
         REFIID iid, void **command_list)
 {
-    struct d3d12_device *device = impl_from_ID3D12Device9(iface);
-    struct d3d12_command_list *object;
-    HRESULT hr;
-
-    TRACE("iface %p, node_mask 0x%08x, type %#x, flags %#x, iid %s, command_list %p.\n",
+    FIXME("iface %p, node_mask 0x%08x, type %#x, flags %#x, iid %s, command_list %p stub!\n",
             iface, node_mask, type, flags, debugstr_guid(iid), command_list);
 
-    if (flags)
-        FIXME("Ignoring flags %#x.\n", flags);
-
-    if (FAILED(hr = d3d12_command_list_create(device, node_mask, type, &object)))
-        return hr;
-
-    return return_interface(&object->ID3D12GraphicsCommandList6_iface,
-            &IID_ID3D12GraphicsCommandList6, iid, command_list);
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d3d12_device_CreateProtectedResourceSession(ID3D12Device9 *iface,
@@ -5209,12 +5193,7 @@ static HRESULT STDMETHODCALLTYPE d3d12_device_EnumerateMetaCommands(ID3D12Device
     FIXME("iface %p, num_meta_commands %p, command_desc %p stub!\n", iface,
             num_meta_commands, command_desc);
 
-    if (!num_meta_commands)
-        return E_INVALIDARG;
-
-    *num_meta_commands = 0;
-
-    return S_OK;
+    return E_NOTIMPL;
 }
 
 static HRESULT STDMETHODCALLTYPE d3d12_device_EnumerateMetaCommandParameters(ID3D12Device9 *iface,
