diff --git a/libs/vkd3d/libs/vkd3d/vkd3d_main.c b/libs/vkd3d/libs/vkd3d/vkd3d_main.c
index 4c58f0a17871f15e4df9d739e5701a3f4bb08e9b..5215cf8ef86eebb733da942c0b7a9e3154080ccf 100644
--- a/libs/vkd3d/libs/vkd3d/vkd3d_main.c
+++ b/libs/vkd3d/libs/vkd3d/vkd3d_main.c
@@ -18,6 +18,8 @@
 
 #include "vkd3d_private.h"
 
+VKD3D_DEBUG_ENV_NAME("VKD3D_DEBUG");
+
 HRESULT vkd3d_create_device(const struct vkd3d_device_create_info *create_info,
         REFIID iid, void **device)
 {
