diff --git a/libs/xml2/include/libxml/parser.h b/libs/xml2/include/libxml/parser.h
index fd6d52074d6e4c07cccc70efe622851d499bba3a..87aacef910fa00d7ddf5ae9e379b5454ba5db2f8 100644
--- a/libs/xml2/include/libxml/parser.h
+++ b/libs/xml2/include/libxml/parser.h
@@ -953,7 +953,7 @@ XML_DEPRECATED XMLPUBFUN int
 		xmlSubstituteEntitiesDefault(int val);
 XML_DEPRECATED XMLPUBFUN int
                 xmlThrDefSubstituteEntitiesDefaultValue(int v);
-XMLPUBFUN int
+XML_DEPRECATED XMLPUBFUN int
 		xmlKeepBlanksDefault	(int val);
 XML_DEPRECATED XMLPUBFUN int
 		xmlThrDefKeepBlanksDefaultValue(int v);
