diff --git a/libs/xml2/include/libxml/xmlversion.h b/libs/xml2/include/libxml/xmlversion.h
index 74e77b376f7cd60ef02c473ad2e178899916099a..2374eaf1aea7aab20cf2662ebde10681efd7cd7d 100644
--- a/libs/xml2/include/libxml/xmlversion.h
+++ b/libs/xml2/include/libxml/xmlversion.h
@@ -29,21 +29,21 @@ XMLPUBFUN void xmlCheckVersion(int version);
  *
  * the version string like "1.2.3"
  */
-#define LIBXML_DOTTED_VERSION "2.12.10"
+#define LIBXML_DOTTED_VERSION "2.12.8"
 
 /**
  * LIBXML_VERSION:
  *
  * the version number: 1.2.3 value is 10203
  */
-#define LIBXML_VERSION 21210
+#define LIBXML_VERSION 21208
 
 /**
  * LIBXML_VERSION_STRING:
  *
  * the version number string, 1.2.3 value is "10203"
  */
-#define LIBXML_VERSION_STRING "21210"
+#define LIBXML_VERSION_STRING "21208"
 
 /**
  * LIBXML_VERSION_EXTRA:
@@ -58,7 +58,7 @@ XMLPUBFUN void xmlCheckVersion(int version);
  * Macro to check that the libxml version in use is compatible with
  * the version the software has been compiled against
  */
-#define LIBXML_TEST_VERSION xmlCheckVersion(21210);
+#define LIBXML_TEST_VERSION xmlCheckVersion(21208);
 
 #ifndef VMS
 #if 0
