diff --git a/libs/xml2/valid.c b/libs/xml2/valid.c
index ea3a65faac8aa2f94e818344d864d9bf756c78ff..886485d67ea845cf0987658b7fa21b67ce7b0b7f 100644
--- a/libs/xml2/valid.c
+++ b/libs/xml2/valid.c
@@ -5057,26 +5057,25 @@ xmlSnprintfElements(char *buf, int size, xmlNodePtr node, int glob) {
 	    return;
 	}
         switch (cur->type) {
-            case XML_ELEMENT_NODE: {
-                int qnameLen = xmlStrlen(cur->name);
-
-                if ((cur->ns != NULL) && (cur->ns->prefix != NULL))
-                    qnameLen += xmlStrlen(cur->ns->prefix) + 1;
-                if (size - len < qnameLen + 10) {
-                    if ((size - len > 4) && (buf[len - 1] != '.'))
-                        strcat(buf, " ...");
-                    return;
-                }
+            case XML_ELEMENT_NODE:
 		if ((cur->ns != NULL) && (cur->ns->prefix != NULL)) {
+		    if (size - len < xmlStrlen(cur->ns->prefix) + 10) {
+			if ((size - len > 4) && (buf[len - 1] != '.'))
+			    strcat(buf, " ...");
+			return;
+		    }
 		    strcat(buf, (char *) cur->ns->prefix);
 		    strcat(buf, ":");
 		}
-                if (cur->name != NULL)
-	            strcat(buf, (char *) cur->name);
+                if (size - len < xmlStrlen(cur->name) + 10) {
+		    if ((size - len > 4) && (buf[len - 1] != '.'))
+			strcat(buf, " ...");
+		    return;
+		}
+	        strcat(buf, (char *) cur->name);
 		if (cur->next != NULL)
 		    strcat(buf, " ");
 		break;
-            }
             case XML_TEXT_NODE:
 		if (xmlIsBlankNode(cur))
 		    break;
