diff --git a/libs/xml2/xmlschemas.c b/libs/xml2/xmlschemas.c
index 76cf68196874f9c00eeaae0d109e9709053ab833..886ff6e8c58a2f733b1ea3ee474c6b11661dbe7f 100644
--- a/libs/xml2/xmlschemas.c
+++ b/libs/xml2/xmlschemas.c
@@ -23390,7 +23390,6 @@ xmlSchemaIDCFillNodeTables(xmlSchemaValidCtxtPtr vctxt,
 			}
 			if (xmlSchemaItemListAdd(bind->dupls, bind->nodeTable[j]) == -1)
 			    goto internal_error;
-                        dupls = (xmlSchemaPSVIIDCNodePtr *) bind->dupls->items;
 			/*
 			* Remove the duplicate entry from the IDC node-table.
 			*/
@@ -23607,8 +23606,6 @@ xmlSchemaBubbleIDCNodeTables(xmlSchemaValidCtxtPtr vctxt)
 				goto internal_error;
 			}
 			xmlSchemaItemListAdd(parBind->dupls, parNode);
-		        dupls = (xmlSchemaPSVIIDCNodePtr *)
-                            parBind->dupls->items;
 		    } else {
 			/*
 			* Add the node-table entry (node and key-sequence) of
