diff --git a/libs/xml2/xpath.c b/libs/xml2/xpath.c
index 660d70e50fdcff8b5d7be402d7ab41cdbfc8541d..83e065262260d1d8ae774d03ab058b427bdf4f35 100644
--- a/libs/xml2/xpath.c
+++ b/libs/xml2/xpath.c
@@ -2644,9 +2644,7 @@ xmlXPathFormatNumber(double number, char buffer[], int buffersize)
 	    /* Finally copy result back to caller */
 	    size = strlen(work) + 1;
 	    if (size > buffersize) {
-#if 0  /* avoid compiler warning */
 		work[buffersize - 1] = 0;
-#endif
 		size = buffersize;
 	    }
 	    memmove(buffer, work, size);
@@ -10714,9 +10712,8 @@ xmlXPathCompLocationPath(xmlXPathParserContextPtr ctxt) {
 	    } else if (CUR == '/') {
 		NEXT;
 		SKIP_BLANKS;
-		if ((CUR != 0) &&
-		    ((IS_ASCII_LETTER(CUR)) || (CUR >= 0x80) ||
-                     (CUR == '_') || (CUR == '.') ||
+		if ((CUR != 0 ) &&
+		    ((IS_ASCII_LETTER(CUR)) || (CUR == '_') || (CUR == '.') ||
 		     (CUR == '@') || (CUR == '*')))
 		    xmlXPathCompRelativeLocationPath(ctxt);
 	    }
