diff --git a/libs/xslt/libxslt/libxslt.h b/libs/xslt/libxslt/libxslt.h
index 5a9ad48c2f4a35f2bb7fc36a7abf6a7712b74d94..5e0d0a45511475dca6756cfc0746395af34ebf49 100644
--- a/libs/xslt/libxslt/libxslt.h
+++ b/libs/xslt/libxslt/libxslt.h
@@ -10,19 +10,6 @@
 #ifndef __XSLT_LIBXSLT_H__
 #define __XSLT_LIBXSLT_H__
 
-/*
- * These macros must be defined before including system headers.
- * Do not add any #include directives above this block.
- */
-#ifndef NO_LARGEFILE_SOURCE
-  #ifndef _LARGEFILE_SOURCE
-    #define _LARGEFILE_SOURCE
-  #endif
-  #ifndef _FILE_OFFSET_BITS
-    #define _FILE_OFFSET_BITS 64
-  #endif
-#endif
-
 #ifdef _WIN32
 #include <win32config.h>
 #else
