diff --git a/libs/xslt/libxslt/numbers.c b/libs/xslt/libxslt/numbers.c
index 8e528381fd75818b0c2f6e9f0bfad093aaf92268..3cd881e3c8d5f8f41a50412d68a6e4f5bb39b91f 100644
--- a/libs/xslt/libxslt/numbers.c
+++ b/libs/xslt/libxslt/numbers.c
@@ -58,46 +58,17 @@ struct _xsltFormat {
     xmlChar		*end;
 };
 
-static const char alpha_upper_list[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
-static const char alpha_lower_list[] = "abcdefghijklmnopqrstuvwxyz";
-static const xsltFormatToken default_token = {
-    BAD_CAST(DEFAULT_SEPARATOR),
-    DEFAULT_TOKEN,
-    1
-};
+static char alpha_upper_list[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ";
+static char alpha_lower_list[] = "abcdefghijklmnopqrstuvwxyz";
+static xsltFormatToken default_token;
 
 /*
- * Helper functions copied from libxml2
- */
-
-/**
- * xsltCopyCharMultiByte:
- * @out:  pointer to an array of xmlChar
- * @val:  the char value
+ * **** Start temp insert ****
  *
- * append the char value in the array
- *
- * Returns the number of xmlChar written
+ * The following routine xsltUTF8Charcmp will be replaced with calls to
+ * the corresponding libxml routine at a later date (when other
+ * inter-library dependencies require it).
  */
-static int
-xsltCopyCharMultiByte(xmlChar *out, int val) {
-    if ((out == NULL) || (val < 0)) return(0);
-    if  (val >= 0x80) {
-	xmlChar *savedout = out;
-	int bits;
-	if (val <   0x800) { *out++= (val >>  6) | 0xC0;  bits=  0; }
-	else if (val < 0x10000) { *out++= (val >> 12) | 0xE0;  bits=  6;}
-	else if (val < 0x110000)  { *out++= (val >> 18) | 0xF0;  bits=  12; }
-	else {
-	    return(0);
-	}
-	for ( ; bits >= 0; bits-= 6)
-	    *out++= ((val >> bits) & 0x3F) | 0x80 ;
-	return (out - savedout);
-    }
-    *out = val;
-    return 1;
-}
 
 /**
  * xsltUTF8Charcmp
@@ -127,6 +98,7 @@ xsltIsLetterDigit(int val) {
            xmlIsDigitQ(val);
 }
 
+/***** Stop temp insert *****/
 /************************************************************************
  *									*
  *			Utility functions				*
@@ -202,7 +174,7 @@ xsltNumberFormatDecimal(xmlBufferPtr buffer,
 		break;
 	    }
 	    pointer -= groupingCharacterLen;
-	    xsltCopyCharMultiByte(pointer, groupingCharacter);
+	    xmlCopyCharMultiByte(pointer, groupingCharacter);
 	}
 
 	val = digit_zero + (int)fmod(number, 10.0);
@@ -220,7 +192,7 @@ xsltNumberFormatDecimal(xmlBufferPtr buffer,
 	 * it is.  So, we generate it into the buffer temp_char, then
 	 * copy from there into temp_string.
 	 */
-	    len = xsltCopyCharMultiByte(temp_char, val);
+	    len = xmlCopyCharMultiByte(temp_char, val);
 	    if ( (pointer - len) < temp_string ) {
 	        i = -1;
 		break;
@@ -246,7 +218,7 @@ xsltNumberFormatAlpha(xsltNumberDataPtr data,
     char temp_string[sizeof(double) * CHAR_BIT * sizeof(xmlChar) + 1];
     char *pointer;
     int i;
-    const char *alpha_list;
+    char *alpha_list;
     double alpha_size = (double)(sizeof(alpha_upper_list) - 1);
 
     /*
@@ -368,6 +340,11 @@ xsltNumberFormatTokenize(const xmlChar *format,
     int val;
     int len;
 
+    default_token.token = DEFAULT_TOKEN;
+    default_token.width = 1;
+    default_token.separator = BAD_CAST(DEFAULT_SEPARATOR);
+
+
     tokens->start = NULL;
     tokens->tokens[0].separator = NULL;
     tokens->end = NULL;
@@ -471,7 +448,7 @@ xsltNumberFormatInsertNumbers(xsltNumberDataPtr data,
 {
     int i = 0;
     double number;
-    const xsltFormatToken *token;
+    xsltFormatTokenPtr token;
 
     /*
      * Handle initial non-alphanumeric token
@@ -679,48 +656,51 @@ xsltNumberFormatGetMultipleLevel(xsltTransformContextPtr context,
 {
     int amount = 0;
     int cnt;
+    xmlNodePtr oldCtxtNode;
     xmlNodePtr ancestor;
     xmlNodePtr preceding;
+    xmlXPathParserContextPtr parser;
+
+    oldCtxtNode = context->xpathCtxt->node;
+    parser = xmlXPathNewParserContext(NULL, context->xpathCtxt);
+    if (parser) {
+	/* ancestor-or-self::*[count] */
+	ancestor = node;
+	while ((ancestor != NULL) && (ancestor->type != XML_DOCUMENT_NODE)) {
+	    if ((fromPat != NULL) &&
+		xsltTestCompMatchList(context, ancestor, fromPat))
+		break; /* for */
 
-    /* ancestor-or-self::*[count] */
-    ancestor = node;
-
-    while ((ancestor != NULL) && (ancestor->type != XML_DOCUMENT_NODE)) {
-        if ((fromPat != NULL) &&
-            xsltTestCompMatchList(context, ancestor, fromPat))
-            break;
-
-        if (xsltTestCompMatchCount(context, ancestor, countPat, node)) {
-            /* count(preceding-sibling::*) */
-            cnt = 1;
-            if (ancestor->type != XML_NAMESPACE_DECL)
-                preceding = ancestor->prev;
-            else
-                preceding = NULL;
-            while (preceding != NULL) {
-                if (xsltTestCompMatchCount(context, preceding, countPat,
-                                           node))
-                    cnt++;
-                preceding = preceding->prev;
-            }
-            array[amount++] = (double)cnt;
-            if (amount >= max)
-                break;
-        }
-
-        if ((ancestor != NULL) && (ancestor->type == XML_NAMESPACE_DECL)) {
-            xmlNsPtr ns = (xmlNsPtr) ancestor;
+            /*
+             * The xmlXPathNext* iterators require that the context node is
+             * set to the start node. Calls to xsltTestCompMatch* may also
+             * leave the context node in an undefined state, so make sure
+             * that the context node is reset before each iterator invocation.
+             */
 
-            if ((ns->next != NULL) &&
-                (ns->next->type != XML_NAMESPACE_DECL))
-                ancestor = (xmlNodePtr) ns->next;
-            else
-                ancestor = NULL;
-        } else {
-            ancestor = ancestor->parent;
-        }
+	    if (xsltTestCompMatchCount(context, ancestor, countPat, node)) {
+		/* count(preceding-sibling::*) */
+		cnt = 1;
+                context->xpathCtxt->node = ancestor;
+                preceding = xmlXPathNextPrecedingSibling(parser, ancestor);
+                while (preceding != NULL) {
+	            if (xsltTestCompMatchCount(context, preceding, countPat,
+                                               node))
+			cnt++;
+                    context->xpathCtxt->node = ancestor;
+                    preceding =
+                        xmlXPathNextPrecedingSibling(parser, preceding);
+		}
+		array[amount++] = (double)cnt;
+		if (amount >= max)
+		    break; /* for */
+	    }
+            context->xpathCtxt->node = node;
+            ancestor = xmlXPathNextAncestor(parser, ancestor);
+	}
+	xmlXPathFreeParserContext(parser);
     }
-
+    context->xpathCtxt->node = oldCtxtNode;
     return amount;
 }
 
@@ -733,12 +713,9 @@ xsltNumberFormatGetValue(xmlXPathContextPtr context,
     int amount = 0;
     xmlBufferPtr pattern;
     xmlXPathObjectPtr obj;
-    xmlNodePtr oldNode;
 
     pattern = xmlBufferCreate();
     if (pattern != NULL) {
-        oldNode = context->node;
-
 	xmlBufferCCat(pattern, "number(");
 	xmlBufferCat(pattern, value);
 	xmlBufferCCat(pattern, ")");
@@ -751,8 +728,6 @@ xsltNumberFormatGetValue(xmlXPathContextPtr context,
 	    xmlXPathFreeObject(obj);
 	}
 	xmlBufferFree(pattern);
-
-        context->node = oldNode;
     }
     return amount;
 }
@@ -1409,3 +1384,4 @@ OUTPUT_NUMBER:
     xmlBufferFree(buffer);
     return status;
 }
+
