diff --git a/dlls/ntdll/unix/process.c b/dlls/ntdll/unix/process.c
--- a/dlls/ntdll/unix/process.c
+++ b/dlls/ntdll/unix/process.c
@@ -1009,6 +1009,11 @@ void fill_vm_counters( VM_COUNTERS_EX *pvmi, int unix_pid )
 #endif
 }
 
+static int get_unix_debugger_pid(void)
+{
+    return 0;
+}
+
 #elif defined(linux)
 
 void fill_vm_counters( VM_COUNTERS_EX *pvmi, int unix_pid )
@@ -1044,6 +1049,23 @@ void fill_vm_counters( VM_COUNTERS_EX *pvmi, int unix_pid )
     fclose(f);
 }
 
+static int get_unix_debugger_pid(void)
+{
+    int unix_pid = 0;
+    char line[256];
+    FILE *file;
+
+    if (!(file = fopen( "/proc/self/status", "r" ))) return 0;
+    while (fgets( line, sizeof(line), file ))
+    {
+        if (sscanf( line, "TracerPid: %d", &unix_pid )) break;
+        unix_pid = 0;
+    }
+    fclose( file );
+
+    return unix_pid;
+}
+
 #elif defined(HAVE_LIBPROCSTAT)
 
 void fill_vm_counters( VM_COUNTERS_EX *pvmi, int unix_pid )
@@ -1068,6 +1090,11 @@ void fill_vm_counters( VM_COUNTERS_EX *pvmi, int unix_pid )
     }
 }
 
+static int get_unix_debugger_pid(void)
+{
+    return 0;
+}
+
 #else
 
 void fill_vm_counters( VM_COUNTERS_EX *pvmi, int unix_pid )
@@ -1075,6 +1102,11 @@ void fill_vm_counters( VM_COUNTERS_EX *pvmi, int unix_pid )
     /* FIXME : real data */
 }
 
+static int get_unix_debugger_pid(void)
+{
+    return 0;
+}
+
 #endif
 
 static BOOL set_hardware_tso( BOOL enable ) {
@@ -1600,6 +1632,14 @@ NTSTATUS WINAPI NtQueryInformationProcess( HANDLE handle, PROCESSINFOCLASS class
         else ret = STATUS_INFO_LENGTH_MISMATCH;
         break;
 
+    case ProcessWineUnixDebuggerPid:
+        len = sizeof(int);
+        if (!info) ret = STATUS_ACCESS_VIOLATION;
+        else if (handle != NtCurrentProcess()) ret = STATUS_INVALID_PARAMETER;
+        else if (size != len) ret = STATUS_INFO_LENGTH_MISMATCH;
+        else *(int *)info = get_unix_debugger_pid();
+        break;
+
     case ProcessQuotaLimits:
         {
             QUOTA_LIMITS qlimits;
