diff --git a/dlls/win32u/sysparams.c b/dlls/win32u/sysparams.c
--- a/dlls/win32u/sysparams.c
+++ b/dlls/win32u/sysparams.c
@@ -332,8 +332,8 @@ static HANDLE get_display_device_init_mutex( void )
     char buffer[256];
     HANDLE mutex;
 
-    snprintf( buffer, ARRAY_SIZE(buffer), "\\Sessions\\%u\\BaseNamedObjects\\display_device_init",
-              NtCurrentTeb()->Peb->SessionId );
+    snprintf( buffer, ARRAY_SIZE(buffer), "\\Sessions\\%u\\BaseNamedObjects\\display_device_init",
+              (int)NtCurrentTeb()->Peb->SessionId );
     name.MaximumLength = asciiz_to_unicode( bufferW, buffer );
     name.Length = name.MaximumLength - sizeof(WCHAR);
 
@@ -450,9 +450,9 @@ static const char *debugstr_devmodew( const DEVMODEW *devmode )
 {
     char position[32] = {0};
     if (devmode->dmFields & DM_POSITION) snprintf( position, sizeof(position), " at %s", wine_dbgstr_point( (POINT *)&devmode->dmPosition ) );
-    return wine_dbg_sprintf( "%ux%u %ubits %uHz rotated %u degrees %sstretched %sinterlaced%s",
-                             devmode->dmPelsWidth, devmode->dmPelsHeight, devmode->dmBitsPerPel,
-                             devmode->dmDisplayFrequency, devmode->dmDisplayOrientation * 90,
+    return wine_dbg_sprintf( "%ux%u %ubits %uHz rotated %u degrees %sstretched %sinterlaced%s",
+                             (UINT)devmode->dmPelsWidth, (UINT)devmode->dmPelsHeight, (UINT)devmode->dmBitsPerPel,
+                             (UINT)devmode->dmDisplayFrequency, (UINT)devmode->dmDisplayOrientation * 90,
                              devmode->dmDisplayFixedOutput == DMDFO_STRETCH ? "" : "un",
                              devmode->dmDisplayFlags & DM_INTERLACED ? "" : "non-",
                              position );
