diff --git a/dlls/win32u/window.c b/dlls/win32u/window.c
--- a/dlls/win32u/window.c
+++ b/dlls/win32u/window.c
@@ -2945,8 +2945,8 @@ BOOL WINAPI NtUserGetWindowPlacement( HWND hwnd, WINDOWPLACEMENT *placement )
     release_win_ptr( win );
 
     TRACE( "%p: returning min %d,%d max %d,%d normal %s\n",
-           hwnd, placement->ptMinPosition.x, placement->ptMinPosition.y,
-           placement->ptMaxPosition.x, placement->ptMaxPosition.y,
+           hwnd, (int)placement->ptMinPosition.x, (int)placement->ptMinPosition.y,
+           (int)placement->ptMaxPosition.x, (int)placement->ptMaxPosition.y,
            wine_dbgstr_rect(&placement->rcNormalPosition) );
     return TRUE;
 }
