diff --git a/programs/cmd/cmd.rc b/programs/cmd/cmd.rc
index bc738faa5e92bdf30a746b421e09ed16ec92c8f3..90091090e113f12e211252d66b7b952862660091 100644
--- a/programs/cmd/cmd.rc
+++ b/programs/cmd/cmd.rc
@@ -406,6 +406,4 @@ Enter HELP <command> for further information on any of the above commands.\n"
   WCMD_BADTOKEN, "Syntax error: unexpected %1\n"
   WCMD_ENDOFLINE, "End of line"
   WCMD_ENDOFFILE, "End of file"
-  WCMD_NUMCOPIED, "%t%1!u! file(s) copied\n"
-  WCMD_NOCOPYTOSELF, "File cannot be copied onto itself.\n"
 }
