diff --git a/programs/cmd/tests/batch.c b/programs/cmd/tests/batch.c
index c4421aa3f6ce3812130aa26ba653196aadf05052..6bc5e916ea94c7f1933265fe0c9877edb9355fe9 100644
--- a/programs/cmd/tests/batch.c
+++ b/programs/cmd/tests/batch.c
@@ -31,38 +31,11 @@ static DWORD path_len;
 static char shortpath[MAX_PATH];
 static DWORD shortpath_len;
 
-static BOOL parse_hexadecimal(const char *p, char *dest)
-{
-    unsigned char c;
-    if (*p++ != '@')  return FALSE;
-    if (*p++ != '\\') return FALSE;
-    if (*p++ != 'x')  return FALSE;
-
-    if (*p >= '0' && *p <= '9') c = *p - '0';
-    else if (*p >= 'a' && *p <= 'f') c = *p - 'a' + 10;
-    else if (*p >= 'A' && *p <= 'F') c = *p - 'A' + 10;
-    else return FALSE;
-    p++;
-
-    c <<= 4;
-
-    if (*p >= '0' && *p <= '9') c += *p - '0';
-    else if (*p >= 'a' && *p <= 'f') c += *p - 'a' + 10;
-    else if (*p >= 'A' && *p <= 'F') c += *p - 'A' + 10;
-    else return FALSE;
-    p++;
-
-    if (*p != '@') return FALSE;
-    *dest = (char)c;
-    return TRUE;
-}
-
 /* Convert to DOS line endings, and substitute escaped whitespace chars with real ones */
 static const char* convert_input_data(const char *data, DWORD size, DWORD *new_size)
 {
     static const char escaped_space[] = {'@','s','p','a','c','e','@'};
     static const char escaped_tab[]   = {'@','t','a','b','@'};
-    static const char escaped_hexadecimal[] = {'@','\\','x','.','.','@'};
     DWORD i, eol_count = 0;
     char *ptr, *new_data;
 
@@ -87,10 +60,6 @@ static const char* convert_input_data(const char *data, DWORD size, DWORD *new_s
                         && !memcmp(data + i, escaped_tab, sizeof(escaped_tab))) {
                     *ptr++ = '\t';
                     i += sizeof(escaped_tab) - 1;
-                } else if (data + i + sizeof(escaped_hexadecimal) - 1 < data + size
-                        && parse_hexadecimal(data + i, ptr)) {
-                    ptr++;
-                    i += sizeof(escaped_hexadecimal) - 1;
                 } else {
                     *ptr++ = data[i];
                 }
@@ -207,7 +176,6 @@ static const char *compare_line(const char *out_line, const char *out_end, const
 {
     const char *out_ptr = out_line, *exp_ptr = exp_line;
     const char *err = NULL;
-    char ch;
 
     static const char pwd_cmd[] = {'@','p','w','d','@'};
     static const char drive_cmd[] = {'@','d','r','i','v','e','@'};
@@ -217,7 +185,6 @@ static const char *compare_line(const char *out_line, const char *out_end, const
     static const char spaces_cmd[] = {'@','s','p','a','c','e','s','@'};
     static const char tab_cmd[]   = {'@','t','a','b','@'};
     static const char formfeed_cmd[]   = {'@','f','o','r','m', 'f', 'e', 'e', 'd', '@'};
-    static const char hexadecimal_cmd[] = {'@','\\','x','0','0', '@'};
     static const char or_broken_cmd[] = {'@','o','r','_','b','r','o','k','e','n','@'};
 
     while(exp_ptr < exp_end) {
@@ -302,15 +269,6 @@ static const char *compare_line(const char *out_line, const char *out_end, const
                 } else {
                     err = out_end;
                 }
-            }else if(exp_ptr+sizeof(hexadecimal_cmd) <= exp_end &&
-                     parse_hexadecimal(exp_ptr, &ch)) {
-                exp_ptr += sizeof(hexadecimal_cmd);
-                if(out_ptr < out_end && *out_ptr == ch) {
-                    out_ptr++;
-                    continue;
-                } else {
-                    err = out_end;
-                }
             }else if(exp_ptr+sizeof(or_broken_cmd) <= exp_end
                      && !memcmp(exp_ptr, or_broken_cmd, sizeof(or_broken_cmd))) {
                 if(out_ptr == out_end)
@@ -524,6 +482,24 @@ static int cmd_available(void)
     return FALSE;
 }
 
+void create_nul_test_file(void)
+{
+    HANDLE file;
+    DWORD size;
+    BOOL bres;
+    char contents[] = "a b c\nd e\0f\ng h i";
+
+    file = CreateFileA("nul_test_file", GENERIC_WRITE, 0, NULL, CREATE_ALWAYS,
+            FILE_ATTRIBUTE_NORMAL, NULL);
+    ok(file != INVALID_HANDLE_VALUE, "CreateFile failed\n");
+    if(file == INVALID_HANDLE_VALUE)
+        return;
+
+    bres = WriteFile(file, contents, ARRAYSIZE(contents), &size, NULL);
+    ok(bres, "Could not write to file: %lu\n", GetLastError());
+    CloseHandle(file);
+}
+
 START_TEST(batch)
 {
     int argc;
@@ -548,9 +524,13 @@ START_TEST(batch)
     }
     shortpath_len = GetShortPathNameA(path, shortpath, ARRAY_SIZE(shortpath));
 
+    create_nul_test_file();
+
     argc = winetest_get_mainargs(&argv);
     if(argc > 2)
         run_from_file(argv[2]);
     else
         EnumResourceNamesA(NULL, "TESTCMD", test_enum_proc, 0);
+
+    DeleteFileA("nul_test_file");
 }
