diff --git a/programs/cmd/tests/test_builtins.bat.exp b/programs/cmd/tests/test_builtins.bat.exp
index 397b720c3bf480043b010a9d0a91546fe12e59af..b26815a9cf25d7be08c0c01d6d587d44679d69e6 100644
--- a/programs/cmd/tests/test_builtins.bat.exp
+++ b/programs/cmd/tests/test_builtins.bat.exp
@@ -77,12 +77,6 @@ SUCCESS 0
 FAILURE 1
 FAILURE 1
 FAILURE 1
-FAILURE 1
-SUCCESS 0
-a@space@
-b@space@
-SUCCESS 0
-a@space@
 --- success/failure for MOVE command
 FAILURE 1
 SUCCESS 0
@@ -100,7 +94,6 @@ FAILURE 1
 SUCCESS 0
 FAILURE 1
 FAILURE 1
-SUCCESS 0
 --- success/failure for change drive command
 SUCCESS 666
 FAILURE 1
@@ -233,10 +226,10 @@ FAILURE 2
 FAILURE 1
 --- success/failure for MORE command
 SUCCESS 0
+SUCCESS 0
 foo@space@
 
 SUCCESS 0
-@todo_wine@SUCCESS 0
-@todo_wine@--- success/failure for PAUSE command
+--- success/failure for PAUSE command
 FAILURE 1
 ---
