diff --git a/programs/cmd/tests/test_builtins.cmd.exp b/programs/cmd/tests/test_builtins.cmd.exp
index d169066a167d9526e8f9fe9f26b72dad925ed2eb..1b5c0c6209c1e08240aa411ae68a6f37a043efaf 100644
--- a/programs/cmd/tests/test_builtins.cmd.exp
+++ b/programs/cmd/tests/test_builtins.cmd.exp
@@ -88,32 +88,13 @@ word
 @tab@word
 
 @pwd@>echo @tab@ on @space@@space@
-
-@todo_wine@@pwd@>echo a 1>nul@space@
-
-@pwd@>if 1 == 2 then @echo a@space@
-
-@todo_wine@@pwd@>IF 1 == 2 ThEn @EchO a@space@
-
-@pwd@>echo 1@space@
-1
-
-@pwd@>echo 2@space@
-2
----
-AA@\x08@BB
 --- @ with chains and brackets
 
 @todo_wine@@pwd@>(echo the @ character chains until  && )  && echo and can hide brackets  || ()  ||@space@
-the @ character chains until
-we leave the current depth
-and can hide brackets
+@todo_wine@the @ character chains until
+@todo_wine@we leave the current depth
+@todo_wine@and can hide brackets
 @todo_wine@---
-foo1
-
-@todo_wine@@pwd@>echo a  1>nul  &&@space@
-foo2
----
 noecho1
 noecho2
 echo3
@@ -146,7 +127,6 @@ word
 @tab@word
 @space@@tab@word
 @tab@word
-3
 ------------ Testing mixed echo modes ------------
 @echo on
 if 1==1 echo foo
@@ -155,22 +135,13 @@ if 1==1 @echo bar
 if 1==1 echo foo2
 if 1==1 @echo bar2
 
-@pwd@>if 1 == 1 echo foo@space@
+@todo_wine@@pwd@>if 1 == 1 echo foo@space@
 foo
 
-@pwd@>if 1 == 1@space@
+@todo_wine@@pwd@>if 1 == 1@space@
 bar
 foo2
 bar2
------------- Testing call and echo modes ------------
-ECHO_IS_ON
-ECHO_IS_OFF
-ECHO_IS_ON
-ECHO_IS_OFF
-ECHO_IS_ON
-ECHO_IS_OFF
-
-@pwd@>rem itsme@space@@space@
 ------------ Testing parameterization ------------
 'a', 'b', 'c'
 '"a b c"', '', ''
@@ -233,9 +204,6 @@ food2
 food21
 foo7@space@@space@@or_broken@not supported@space@
 foo@or_broken@not supported
---- right-to-left redirection
-good
-foo
 --- redirect at beginning of line
 foo
 foo1
@@ -457,7 +425,7 @@ h1 else echo h2
 ---
 i1 else echo i2
 i3
-j2@space@
+@todo_wine@j2@space@
 ---
 k1
 k2
@@ -483,42 +451,6 @@ p2
 ---
 q2
 q3
-------------- Testing for variables expansion in pipes
-a@space@@space@
-b@space@@space@
-
----
-a-c@space@@space@
-a-d@space@@space@
-b-c@space@@space@
-b-d@space@@space@
-
----
-a same a@space@
-a diff b@space@
-b diff a@space@
-b same b@space@
-
----
-foo
-foo
----
-yy!WINE_VAR!yy
-yy!WINE_VAR!yy@space@
-
-yyfooyy
-yyfooyy@space@
-
-yy!WINE_VAR!yy@space@
-
-yy!WINE_VAR!yy@space@
-
-
----
-bar1bar@space@
-
-bar2bar@space@
-
 ------------- Testing internal commands return codes
 --- success/failure for basics
 SUCCESS 0
@@ -596,12 +528,6 @@ SUCCESS 0
 FAILURE 1
 FAILURE 1
 FAILURE 1
-FAILURE 1
-SUCCESS 0
-a@space@
-b@space@
-SUCCESS 0
-a@space@
 --- success/failure for MOVE command
 FAILURE 1
 SUCCESS 0
@@ -619,7 +545,6 @@ FAILURE 1
 SUCCESS 0
 FAILURE 1
 FAILURE 1
-SUCCESS 0
 --- success/failure for change drive command
 SUCCESS 666
 FAILURE 1
@@ -752,49 +677,14 @@ FAILURE 2
 FAILURE 1
 --- success/failure for MORE command
 SUCCESS 0
+SUCCESS 0
 foo@space@
 
 SUCCESS 0
-@todo_wine@SUCCESS 0
-@todo_wine@--- success/failure for PAUSE command
+--- success/failure for PAUSE command
 FAILURE 1
 ---
---------- success/failure when invoking cmd /c --------------
-a
-SUCCESS 0
-FAILURE 1
-FAILURE 1
-FAILURE 457
-SUCCESS 0
-FAILURE 459
-FAILURE 460
-FAILURE 461
-FAILURE 267
-FAILURE 1
-FAILURE 1
---------- success/failure when invoking cmd /k --------------
-a
-ERRORLEVEL 0
-SUCCESS 0
-ERRORLEVEL 1
-SUCCESS 0
-ERRORLEVEL 1
-SUCCESS 0
-FAILURE 457
-ERRORLEVEL 458
-SUCCESS 0
-ERRORLEVEL 459
-SUCCESS 0
-ERRORLEVEL 460
-SUCCESS 0
-ERRORLEVEL 461
-SUCCESS 0
-ERRORLEVEL 0
-SUCCESS 0
-ERRORLEVEL 9009
-SUCCESS 0
-@todo_wine@FAILURE 1
-@todo_wine@------------ Testing 'set' ------------
+------------ Testing 'set' ------------
 1
 0
 WINE_FOOBAR not defined
@@ -882,18 +772,6 @@ e@or_broken@qwerty
 ''@or_broken@'qwerty'
 r@or_broken@qwerty
 ty
-'y'
-''
-''
-'qwerty'
-''
-'qw'
-'WINE_VAR:~2a'
-'WINE_VAR:~2a,2'
-'WINE_VAR:~a,2'
-'WINE_VAR:~2,2a'
-'WINE_VAR:~-'
------------- Testing variable partial replacement ------------
 azertyazERTY
 ertyERTY
 =_QWERTY
@@ -994,9 +872,6 @@ gotitright
 foo
 !WINE_FOO!
 not empty
---- nested expansion
-foo
-foo
 --- using /V cmd flag
 foo
 foo@or_broken@!WINE_FOO!
@@ -1060,10 +935,6 @@ foobay
 
 
 ---7
----8
-passed
----9
-passed
 ------------ Testing NUL ------------
 bar
 bar
@@ -1103,7 +974,6 @@ with closing bracket
 255 two lines
 before both blocks
 255 nested
-0 unmatched
 --- case sensitivity with and without /i option
 if seems to default to case sensitivity
 if /i seems to work
@@ -1886,49 +1756,6 @@ bar\baz removed
 @pwd@
 @drive@@path@foobar
 @pwd@
------------- Testing dir /o ------------
----
-a1.ab
-a2.ac
-a1.aa
-a1.ac
-a2.aa
----
-a1.ab
-a2.ac
-a1.aa
-a1.ac
-a2.aa
----
-a2.ac
-a1.ab
-a2.aa
-a1.ac
-a1.aa
----
-a1.ab
-a2.ac
-a1.aa
-a1.ac
-a2.aa
----
-a2.ac
-a1.ab
-a2.aa
-a1.ac
-a1.aa
----
-a2.ac
-a1.ab
-a1.ac
-a2.aa
-a1.aa
----
-a2.ac
-a1.ab
-a2.aa
-a1.ac
-a1.aa
 ------------ Testing attrib ------------
 A@spaces@@drive@@path@foobar\foo@or_broken@A       I@spaces@@drive@@path@foobar\foo
 A@spaces@@drive@@path@foobar\foo@or_broken@A       I@spaces@@drive@@path@foobar\foo
@@ -2002,7 +1829,7 @@ Should expand foobaz
 batfile
 robinfile
 0
-1
+@todo_wine@1
 1
 1
 non-builtin dir
@@ -2052,33 +1879,19 @@ file correctly deleted
 --- a batch file can alter itself
 bar
 ---------- Testing copy
-Passed: Did not find file1 in COPY output
 Passed: Found expected dummy.file
-Passed: Did not find file1 in COPY output
 Passed: Found expected dir1\file1
-Passed: Did not find file1 in COPY output
 Passed: Found expected dir1\file1
-Passed: Did not find file1 in COPY output
 Passed: Found expected dir1\file99
-Passed: Did not find file1 in COPY output
 Passed: Found expected file1
 Passed: Did not find dir2
 Passed: Did not find dir2\file1
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
 Passed: Found expected dir1\file1
 Passed: Found expected dir1\file2
 Passed: Found expected dir1\file3
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
 Passed: Found expected dir1\file1
 Passed: Found expected dir1\file2
 Passed: Found expected dir1\file3
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
 Passed: Found expected file1
 Passed: Found expected file2
 Passed: Found expected file3
@@ -2086,42 +1899,14 @@ Passed: Did not find dir2
 Passed: Did not find dir2\file1
 Passed: Did not find dir2\file2
 Passed: Did not find dir2\file3
-Passed: Did not find file2 in COPY output
 Passed: Found expected testfile
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
 Passed: Found expected testfile
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
 Passed: Found expected testfile
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
-Passed: Found expected fred in COPY output
 Passed: Found expected testfile
-Passed: Found expected fred in COPY output
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
 Passed: Found expected testfile
-Passed: Found expected fred in COPY output
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
 Passed: Found expected dir1\fred
-Passed: Found expected fred in COPY output
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
 Passed: Found expected dir1\fred
-Passed: Found expected fred in COPY output
-Passed: Found expected file1 in COPY output
-Passed: Found expected file2 in COPY output
-Passed: Found expected file3 in COPY output
 Passed: Found expected fred
-Passed: Found expected fred in COPY output
-Passed: Found expected file1 in COPY output
 Passed: Found expected fred
 Passed: file size check on file1 [5]@or_broken@Skipping file size check on NT4
 Passed: file size check on file2 [8]@or_broken@Skipping file size check on NT4
@@ -2238,13 +2023,8 @@ goto with a following space worked
 goto with following amphersands worked
 goto with redirections worked
 Ignoring double colons worked
-dest7 1
-dest8 1
 label with mixed whitespace and no echo worked
 Correctly ignored trailing information
-desteof1 666
-desteof2 666
----
 Begin:
 ..First sub
 ..First sub
@@ -2285,19 +2065,6 @@ Normal+tab+garbage
 Success
 @todo_wine@Success
 ---- Testing nasty bits ----
------------- Testing updated code page execution ------------
-Success
------------- Testing alteration while executing ------
-BB@space@
----
-BB@space@
----
-A2@space@
-BB@space@
----
-A2@space@
-BB@space@
----
 ------------ Testing combined CALLs/GOTOs ------------
 world
 cheball
