diff --git a/programs/cmd/wcmd.h b/programs/cmd/wcmd.h
index c8d0d03e9e60f78c3c214455f2ac7dc7e470bb5b..caacd44995d20657db8fbbeb7a69f4ce81db5078 100644
--- a/programs/cmd/wcmd.h
+++ b/programs/cmd/wcmd.h
@@ -56,7 +56,6 @@ typedef enum _CMD_OPERATOR
     CMD_ONFAILURE,   /* ||                      */
     CMD_ONSUCCESS,   /* &&                      */
     CMD_PIPE,        /* Single |                */
-    CMD_BLOCK,       /* ( block )               */
     CMD_IF,          /* IF command              */
     CMD_FOR,         /* FOR command             */
 } CMD_OPERATOR;
@@ -112,7 +111,6 @@ typedef struct _CMD_FOR_CONTROL
 typedef struct _CMD_NODE
 {
     CMD_OPERATOR      op;            /* operator */
-    BOOL              do_echo;
     CMD_REDIRECTION  *redirects;     /* Redirections */
     union
     {
@@ -133,10 +131,6 @@ typedef struct _CMD_NODE
             CMD_FOR_CONTROL   for_ctrl;
             struct _CMD_NODE *do_block;
         };
-        struct                       /* CMD_BLOCK */
-        {
-            struct _CMD_NODE *block;
-        };
     };
 } CMD_NODE;
 
@@ -153,14 +147,7 @@ struct _DIRECTORY_STACK *WCMD_dir_stack_free(struct _DIRECTORY_STACK *dir);
 typedef int RETURN_CODE;
 #define RETURN_CODE_SYNTAX_ERROR         255
 #define RETURN_CODE_CANT_LAUNCH          9009
-#define RETURN_CODE_ABORTED              (-999999) /* generated for exit /b so that all loops (and al.) are exited*/
-#define RETURN_CODE_GOTO                 (-999998) /* generated when changing file position (and break from if/for instructions) */
-#define RETURN_CODE_EXITED               (-999997) /* generated when batch file terminates because child has terminated */
-/* to test if one shall break from instruction within a batch file */
-static inline BOOL WCMD_is_break(RETURN_CODE return_code)
-{
-    return return_code == RETURN_CODE_ABORTED || return_code == RETURN_CODE_GOTO;
-}
+#define RETURN_CODE_ABORTED              (-999999)
 
 BOOL WCMD_print_volume_information(const WCHAR *);
 
@@ -184,12 +171,11 @@ RETURN_CODE WCMD_label(void);
 void WCMD_leave_paged_mode(void);
 RETURN_CODE WCMD_more(WCHAR *);
 RETURN_CODE WCMD_move (void);
-WCHAR* WINAPIV WCMD_format_string(const WCHAR *format, ...);
-void WINAPIV WCMD_output(const WCHAR *format, ...);
-void WINAPIV WCMD_output_stderr(const WCHAR *format, ...);
-RETURN_CODE WCMD_output_asis(const WCHAR *message);
-RETURN_CODE WCMD_output_flush(void);
-RETURN_CODE WCMD_output_asis_stderr(const WCHAR *message);
+WCHAR* WINAPIV WCMD_format_string (const WCHAR *format, ...);
+void WINAPIV WCMD_output (const WCHAR *format, ...);
+void WINAPIV WCMD_output_stderr (const WCHAR *format, ...);
+void WCMD_output_asis (const WCHAR *message);
+void WCMD_output_asis_stderr (const WCHAR *message);
 RETURN_CODE WCMD_pause(void);
 RETURN_CODE WCMD_popd(void);
 void WCMD_print_error (void);
@@ -224,13 +210,11 @@ void WCMD_HandleTildeModifiers(WCHAR **start, BOOL atExecute);
 WCHAR *WCMD_strip_quotes(WCHAR *cmd);
 WCHAR *WCMD_LoadMessage(UINT id);
 WCHAR *WCMD_strsubstW(WCHAR *start, const WCHAR* next, const WCHAR* insert, int len);
-RETURN_CODE WCMD_wait_for_input(HANDLE hIn);
-RETURN_CODE WCMD_wait_for_console_input(void);
-BOOL WCMD_read_console(const HANDLE hInput, WCHAR *inputBuffer, const DWORD inputBufferLength, LPDWORD numRead);
+BOOL WCMD_ReadFile(const HANDLE hIn, WCHAR *intoBuf, const DWORD maxChars, LPDWORD charsRead);
 
 enum read_parse_line {RPL_SUCCESS, RPL_EOF, RPL_SYNTAXERROR};
-enum read_parse_line WCMD_ReadAndParseLine(CMD_NODE **output);
-void node_dispose_tree(CMD_NODE *cmds);
+enum read_parse_line WCMD_ReadAndParseLine(const WCHAR *initialcmd, CMD_NODE **output);
+void      node_dispose_tree(CMD_NODE *cmds);
 RETURN_CODE node_execute(CMD_NODE *node);
 
 RETURN_CODE WCMD_call_batch(const WCHAR *, WCHAR *);
@@ -240,8 +224,6 @@ RETURN_CODE WCMD_run_builtin_command(int cmd_index, WCHAR *cmd);
 BOOL WCMD_find_label(HANDLE h, const WCHAR*, LARGE_INTEGER *pos);
 void WCMD_set_label_end(WCHAR *string);
 
-RETURN_CODE WCMD_ctrlc_status(void);
-
 void *xrealloc(void *, size_t) __WINE_ALLOC_SIZE(2) __WINE_DEALLOC(free);
 
 static inline void *xalloc(size_t sz) __WINE_MALLOC;
@@ -272,37 +254,23 @@ static inline BOOL ends_with_backslash( const WCHAR *path )
 
 int evaluate_if_condition(WCHAR *p, WCHAR **command, int *test, int *negate);
 
-/* Data structure to store information about a batch file */
-struct batch_file
-{
-    unsigned            ref_count;     /* number of BATCH_CONTEXT attached to this */
-    WCHAR              *path_name;     /* Name of self */
-    FILETIME            last_modified;
-    struct
-    {
-        LARGE_INTEGER   from;
-        LARGE_INTEGER   position;
-        unsigned        age;
-        const WCHAR    *label;
-    } cache[8];
-};
+/* Data structure to hold context when executing batch files */
 
-struct batch_context
+typedef struct _BATCH_CONTEXT
 {
-    WCHAR                *command;	    /* The command which invoked the batch file */
-    LARGE_INTEGER         file_position;
-    int                   shift_count[10];  /* Offset in terms of shifts for %0 - %9 */
-    struct batch_context *prev_context;     /* Pointer to the previous context block */
-    struct batch_file    *batch_file;       /* Reference to the file itself */
-};
-
-#define WCMD_FILE_POSITION_EOF (~(DWORD64)0)
+    WCHAR *command;	  /* The command which invoked the batch file */
+    LARGE_INTEGER file_position;
+    WCHAR *batchfileW;    /* Name of same */
+    int shift_count[10];  /* Offset in terms of shifts for %0 - %9 */
+    struct _BATCH_CONTEXT *prev_context; /* Pointer to the previous context block */
+    BOOL  skip_rest;      /* Skip the rest of the batch program and exit */
+} BATCH_CONTEXT;
 
 /* Data structure to handle building lists during recursive calls */
 
 struct env_stack
 {
-    struct batch_context *context;
+    BATCH_CONTEXT *context;
     struct env_stack *next;
     union
     {
@@ -352,20 +320,20 @@ void WCMD_set_for_loop_variable(unsigned varidx, const WCHAR *value);
  */
 extern WCHAR quals[MAXSTRING], param1[MAXSTRING], param2[MAXSTRING];
 extern int errorlevel;
-extern struct batch_context *context;
+extern BATCH_CONTEXT *context;
 extern BOOL delayedsubst;
 
 static inline BOOL WCMD_is_in_context(const WCHAR *ext)
 {
     size_t c_len, e_len;
-    if (!context || !context->batch_file) return FALSE;
+    if (!context) return FALSE;
     if (!ext) return TRUE;
-    c_len = wcslen(context->batch_file->path_name);
+    c_len = wcslen(context->batchfileW);
     e_len = wcslen(ext);
-    return (c_len > e_len) && !wcsicmp(&context->batch_file->path_name[c_len - e_len], ext);
+    return (c_len > e_len) && !wcsicmp(&context->batchfileW[c_len - e_len], ext);
 }
 
-#endif /* !RC_INVOKED */
+ #endif /* !RC_INVOKED */
 
 /*
  *	Serial nos of builtin commands. These constants must be in step with
@@ -482,5 +450,3 @@ extern WCHAR version_string[];
 #define WCMD_BADTOKEN         1047
 #define WCMD_ENDOFLINE        1048
 #define WCMD_ENDOFFILE        1049
-#define WCMD_NUMCOPIED        1050
-#define WCMD_NOCOPYTOSELF     1051
