diff --git a/programs/dism/version.rc b/programs/dism/version.rc
index 013d46faf914b679cbed6dd7360739dda2dc49d3..d0021b9b406242f337854539d2748e92539bb1f2 100644
--- a/programs/dism/version.rc
+++ b/programs/dism/version.rc
@@ -20,5 +20,8 @@
 #define WINE_FILENAME_STR "dism.exe"
 #define WINE_FILETYPE VFT_APP
 #define WINE_FILEVERSION 10,0,18362,900
+#define WINE_FILEVERSION_STR "10.0.18362.900"
+#define WINE_PRODUCTVERSION 10,0,18362,900
+#define WINE_PRODUCTVERSION_STR "10.0.18362.900"
 
 #include "wine/wine_common_ver.rc"
