diff --git a/programs/dpnsvr/version.rc b/programs/dpnsvr/version.rc
index 6340d227ec0a0a14ec7ac63f1ed38d50457366ff..c740736243f276c313f4fcd908045602ce53fd99 100644
--- a/programs/dpnsvr/version.rc
+++ b/programs/dpnsvr/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine dpnsvr"
 #define WINE_FILENAME_STR "dpnsvr.dll"
 #define WINE_FILEVERSION 5,3,0,900
+#define WINE_FILEVERSION_STR "5.3.0.900"
+#define WINE_PRODUCTVERSION 5,3,0,900
+#define WINE_PRODUCTVERSION_STR "5.3.0.900"
 
 #include "wine/wine_common_ver.rc"
