diff --git a/programs/dpvsetup/version.rc b/programs/dpvsetup/version.rc
index 086ac8d5cc6f23c2f4a96f4e5bdf77ba942ced1d..ed2ada993f9d91a26d56fc583b2d0b9696b3e6ad 100644
--- a/programs/dpvsetup/version.rc
+++ b/programs/dpvsetup/version.rc
@@ -19,5 +19,8 @@
 #define WINE_FILEDESCRIPTION_STR "Wine dpvsetup"
 #define WINE_FILENAME_STR "dpvsetup.exe"
 #define WINE_FILEVERSION 5,3,2600,5512
+#define WINE_FILEVERSION_STR "5.3.2600.5512"
+#define WINE_PRODUCTVERSION 5,3,2600,5512
+#define WINE_PRODUCTVERSION_STR "5.3.2600.5512"
 
 #include "wine/wine_common_ver.rc"
