diff --git a/programs/dxdiag/main.c b/programs/dxdiag/main.c
index b6ea5306d06a167daaa6007ff6509d43a4632bbb..47909dcf2c637e6d20186b6708726349921412b9 100644
--- a/programs/dxdiag/main.c
+++ b/programs/dxdiag/main.c
@@ -72,7 +72,13 @@ static BOOL process_file_name(const WCHAR *cmdline, enum output_type output_type
         endptr = cmdline + lstrlenW(cmdline);
 
     len = endptr - cmdline;
-    if (len == 0 || len >= filename_len)
+    if (len == 0)
+    {
+        *filename = 0;
+        return TRUE;
+    }
+
+    if (len >= filename_len)
         return FALSE;
 
     memcpy(filename, cmdline, len * sizeof(WCHAR));
