diff --git a/programs/explorer/explorer.c b/programs/explorer/explorer.c
index 185414417c90807d50ecdb0b7a545e866ae342d6..7652abaf653eb3cafacb3312b42a8f77df6bb02c 100644
--- a/programs/explorer/explorer.c
+++ b/programs/explorer/explorer.c
@@ -40,7 +40,7 @@ WINE_DEFAULT_DEBUG_CHANNEL(explorer);
 #define EXPLORER_INFO_INDEX 0
 
 #define NAV_TOOLBAR_HEIGHT 30
-#define PATHBOX_HEIGHT 240
+#define PATHBOX_HEIGHT 24
 static int nav_toolbar_height;
 static int pathbox_height;
 
@@ -862,28 +862,27 @@ static void parse_command_line(LPWSTR commandline,parameters_struct *parameters)
     while (*p)
     {
         while (iswspace(*p)) p++;
-        if (wcsnicmp(p, arg_n, len_n )==0)
+        if (wcsncmp(p, arg_n, len_n )==0)
         {
             parameters->explorer_mode = FALSE;
             p += len_n;
-            if (*p == ',') p++;
         }
-        else if (wcsnicmp(p, arg_e, len_e )==0)
+        else if (wcsncmp(p, arg_e, len_e )==0)
         {
             parameters->explorer_mode = TRUE;
             p += len_e;
         }
-        else if (wcsnicmp(p, arg_cd, len_cd )==0)
+        else if (wcsncmp(p, arg_cd, len_cd )==0)
         {
             p += len_cd;
             p = copy_path_string(parameters->root,p);
         }
-        else if (wcsnicmp(p, arg_root, len_root )==0)
+        else if (wcsncmp(p, arg_root, len_root )==0)
         {
             p += len_root;
             p = copy_path_string(parameters->root,p);
         }
-        else if (wcsnicmp(p, arg_select, len_select )==0)
+        else if (wcsncmp(p, arg_select, len_select )==0)
         {
             p += len_select;
             p = copy_path_string(parameters->selection,p);
@@ -891,13 +890,13 @@ static void parse_command_line(LPWSTR commandline,parameters_struct *parameters)
                 copy_path_root(parameters->root,
                                parameters->selection);
         }
-        else if (wcsnicmp(p, arg_desktop, len_desktop )==0)
+        else if (wcsncmp(p, arg_desktop, len_desktop )==0)
         {
             p += len_desktop;
             manage_desktop( p );  /* the rest of the command line is handled by desktop mode */
         }
         /* workaround for Worms Armageddon that hardcodes a /desktop option with quotes */
-        else if (wcsnicmp(p, arg_desktop_quotes, len_desktop_quotes )==0)
+        else if (wcsncmp(p, arg_desktop_quotes, len_desktop_quotes )==0)
         {
             p += len_desktop_quotes;
             manage_desktop( p );  /* the rest of the command line is handled by desktop mode */
