diff --git a/programs/explorer/explorer.rc b/programs/explorer/explorer.rc
index 72cd9a13a5a69577974043e2f5b1119e2f6ab7b5..e932a6080bd02e6513bb3e67090d83e98f1a5b13 100644
--- a/programs/explorer/explorer.rc
+++ b/programs/explorer/explorer.rc
@@ -42,6 +42,9 @@ LANGUAGE LANG_NEUTRAL, SUBLANG_NEUTRAL
 #define WINE_FILEDESCRIPTION_STR "Wine Explorer"
 #define WINE_FILENAME_STR "explorer.exe"
 #define WINE_FILEVERSION 6,1,7601,17567
+#define WINE_FILEVERSION_STR "6.1.7601.17567"
+#define WINE_PRODUCTVERSION 6,1,7601,17567
+#define WINE_PRODUCTVERSION_STR "6.1.7601.17567"
 
 #include "wine/wine_common_ver.rc"
 
