diff --git a/programs/find/tests/find.c b/programs/find/tests/find.c
index 70d3cb9338e65154ff5c20834583a927f8282436..10fd8ad8ae3e1e36591cd2e16928b9bf218fff0d 100644
--- a/programs/find/tests/find.c
+++ b/programs/find/tests/find.c
@@ -105,7 +105,7 @@ static void mangle_text(const BYTE *input, int input_len, BYTE *output, int outp
     int count_wchar;
 
     /* Check for UTF-16 LE BOM */
-    if (input_len > 1 && input[0] == 0xFF && input[1] == 0xFE)
+    if (input[0] == 0xFF && input[1] == 0xFE)
     {
         int buffer_count = 0;
         int i;
