diff --git a/programs/findstr/tests/findstr.c b/programs/findstr/tests/findstr.c
index e766b542bb5e6e851e518a0e6dbfd023fb7d6bae..f72d9220cf6dfa17e2a632ed9f9ee1905efbb2b7 100644
--- a/programs/findstr/tests/findstr.c
+++ b/programs/findstr/tests/findstr.c
@@ -148,69 +148,69 @@ static void test_basic(void)
     ok(!ret, "Got the wrong result.\n");
 
     /* find string in stdin, LF */
-    run_find_stdin("/L abc", "abc\n", 0);
+    run_find_stdin("abc", "abc\n", 0);
     ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
     ret = strcmp(stdout_buffer, "abc\n");
     ok(!ret, "Got the wrong result.\n");
 
     /* find string in stdin, CR/LF */
-    run_find_stdin("/L abc", "abc\r\n", 0);
+    run_find_stdin("abc", "abc\r\n", 0);
     ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
     ret = strcmp(stdout_buffer, "abc\r\n");
     ok(!ret, "Got the wrong result.\n");
 
     /* find string in stdin fails */
-    run_find_stdin("/L abc", "cba", 1);
+    run_find_stdin("abc", "cba", 1);
     ok(stdout_size == 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
 
     /* find string in file */
-    run_find_file("/L abc", "abc", 0);
+    run_find_file("abc", "abc", 0);
     ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
     ret = strcmp(stdout_buffer, "abc");
     ok(!ret, "Got the wrong result.\n");
 
     /* find string in file fails */
-    run_find_file("/L abc", "cba", 1);
+    run_find_file("abc", "cba", 1);
     ok(stdout_size == 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
 
     /* find string in stdin with space separator */
-    run_find_stdin("/L \"abc cba\"", "abc", 0);
+    run_find_stdin("\"abc cba\"", "abc", 0);
     ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
     ret = strcmp(stdout_buffer, "abc\r\n");
     ok(!ret, "Got the wrong result.\n");
 
     /* find string in stdin with /C: */
-    run_find_stdin("/L /C:abc", "abc", 0);
+    run_find_stdin("/C:abc", "abc", 0);
     ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
     ret = strcmp(stdout_buffer, "abc\r\n");
     ok(!ret, "Got the wrong result.\n");
 
     /* find string in stdin with /C:"abc" */
-    run_find_stdin("/L /C:\"abc\"", "abc", 0);
+    run_find_stdin("/C:\"abc\"", "abc", 0);
     ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
     ret = strcmp(stdout_buffer, "abc\r\n");
     ok(!ret, "Got the wrong result.\n");
 
     /* find string in stdin with /C:"abc cba" fails */
-    run_find_stdin("/L /C:\"abc cba\"", "abc", 1);
+    run_find_stdin("/C:\"abc cba\"", "abc", 1);
     ok(stdout_size == 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
 
     /* find string in stdin with /C: fails */
-    run_find_stdin("/L /C:abc", "cba", 1);
+    run_find_stdin("/C:abc", "cba", 1);
     ok(stdout_size == 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
 
     /* find string in file, case insensitive */
-    run_find_file("/L /I aBc", "abc", 0);
+    run_find_file("/I aBc", "abc", 0);
     ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
     ret = strcmp(stdout_buffer, "abc");
@@ -314,23 +314,6 @@ static void test_basic(void)
     ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
     ret = strcmp(stdout_buffer, "ab$c");
     ok(!ret, "Got the wrong result. '%s'\n", stdout_buffer);
-
-    run_find_file("/R .", "a", 0);
-    ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
-    ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
-    ret = strcmp(stdout_buffer, "a");
-    ok(!ret, "Got the wrong result. '%s'\n", stdout_buffer);
-
-    run_find_file(".", "a", 0);
-    ok(stdout_size > 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
-    ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
-    ret = strcmp(stdout_buffer, "a");
-    ok(!ret, "Got the wrong result. '%s'\n", stdout_buffer);
-
-    run_find_file("/L .", "a", 1);
-    ok(stdout_size == 0, "Unexpected stdout buffer size %ld.\n", stdout_size);
-    ok(stderr_size == 0, "Unexpected stderr buffer size %ld.\n", stderr_size);
-    ok(!ret, "Got the wrong result. '%s'\n", stdout_buffer);
 }
 
 START_TEST(findstr)
