diff --git a/programs/iexplore/iexplore.rc b/programs/iexplore/iexplore.rc
index 8635ec4c8a3a0077c6c836c16d4254e1be1cb5cc..3d6d50fdeccdbf49a69b83e96ab39f860fb9ac43 100644
--- a/programs/iexplore/iexplore.rc
+++ b/programs/iexplore/iexplore.rc
@@ -19,6 +19,9 @@
 #define WINE_FILEDESCRIPTION_STR "Wine Internet Explorer"
 #define WINE_FILENAME_STR "iexplore.exe"
 #define WINE_FILEVERSION 9,0,8112,16421
+#define WINE_FILEVERSION_STR "9.0.8112.16421"
+#define WINE_PRODUCTVERSION 9,0,8112,16421
+#define WINE_PRODUCTVERSION_STR "9.0.8112.16421"
 #define WINE_EXTRAVALUES VALUE "OLESelfRegister",""
 
 #include "wine/wine_common_ver.rc"
