diff --git a/programs/msiexec/rsrc.rc b/programs/msiexec/rsrc.rc
index 22d58c0a653041dd1139d2acdd189938a53ae45b..9cc45a446ee685323336b6a0a61b7ff2825bdd87 100644
--- a/programs/msiexec/rsrc.rc
+++ b/programs/msiexec/rsrc.rc
@@ -22,6 +22,9 @@
 #define WINE_FILENAME_STR "msiexec.exe"
 #define WINE_FILETYPE VFT_APP
 #define WINE_FILEVERSION 4,5,6001,22308
+#define WINE_FILEVERSION_STR "4.5.6001.22308"
+#define WINE_PRODUCTVERSION 4,5,6001,22308
+#define WINE_PRODUCTVERSION_STR "4.5.6001.22308"
 #define WINE_PRODUCTNAME_STR "Wine Installer"
 
 #include "wine/wine_common_ver.rc"
