diff --git a/programs/ngen/version.rc b/programs/ngen/version.rc
index 683447165eb1043216d65a4a2a42511aff766556..a3b004c6eb9bcf7f9207e3e5e34ccf39897de1e8 100644
--- a/programs/ngen/version.rc
+++ b/programs/ngen/version.rc
@@ -20,5 +20,8 @@
 #define WINE_FILENAME_STR "ngen.exe"
 #define WINE_FILETYPE VFT_APP
 #define WINE_FILEVERSION 2,0,50727,1433
+#define WINE_FILEVERSION_STR "2.0.50727.1433"
+#define WINE_PRODUCTVERSION 2,0,50727,1433
+#define WINE_PRODUCTVERSION_STR "2.0.50727.1433"
 
 #include "wine/wine_common_ver.rc"
