diff --git a/programs/regsvr32/regsvr32.c b/programs/regsvr32/regsvr32.c
index 0293c1cc774ef396c3149c8c079b5d93f1e3c19a..b0889c9034abc68ee4272ab058a9a20cbea9e3a9 100644
--- a/programs/regsvr32/regsvr32.c
+++ b/programs/regsvr32/regsvr32.c
@@ -337,8 +337,6 @@ int __cdecl wmain(int argc, WCHAR* argv[])
             case 's':
                 Silent = TRUE;
                 break;
-            case 'c':  /* output to console, ignored in recent Windows */
-                break;
             case 'i':
                 CallInstall = TRUE;
                 wsCommandLine = parse_command_line(argv[i] + 2); /* argv[i] + strlen("/i") */
