diff --git a/programs/robocopy/robocopy.rc b/programs/robocopy/robocopy.rc
index 71589c672367eb89014f7735dfbb9ceebc4b97d0..ffd5f770b1bc577438bf3d1f31a5afafd497920e 100644
--- a/programs/robocopy/robocopy.rc
+++ b/programs/robocopy/robocopy.rc
@@ -26,6 +26,9 @@ LANGUAGE LANG_NEUTRAL, SUBLANG_NEUTRAL
 #define WINE_FILENAME_STR "robocopy.exe"
 #define WINE_FILETYPE VFT_APP
 #define WINE_FILEVERSION 5,1,10,1027
+#define WINE_FILEVERSION_STR "5.1.10.1027"
 
+#define WINE_PRODUCTVERSION 5,1,10,1027
+#define WINE_PRODUCTVERSION_STR "5.1.10.1027"
 
 #include "wine/wine_common_ver.rc"
