diff --git a/programs/services/rpc.c b/programs/services/rpc.c
index b2f541ea2c0039fa63daa67f69b0642388a47724..798f674ecf772706033601dc8c896ffa2bfa6bc0 100644
--- a/programs/services/rpc.c
+++ b/programs/services/rpc.c
@@ -986,10 +986,6 @@ DWORD __cdecl svcctl_ChangeServiceConfig2W( SC_RPC_HANDLE hService, SC_RPC_CONFI
         save_service_config( service->service_entry );
         service_unlock( service->service_entry );
         break;
-    case SERVICE_CONFIG_SERVICE_SID_INFO:
-        WINE_FIXME( "SERVICE_CONFIG_SERVICE_SID_INFO not implemented: type %lu\n",
-                    config.sid->dwServiceSidType );
-        break;
     default:
         WINE_FIXME("level %lu not implemented\n", config.dwInfoLevel);
         err = ERROR_INVALID_LEVEL;
