diff --git a/programs/tasklist/tasklist.rc b/programs/tasklist/tasklist.rc
index 0dac50c7d89b777ef1d3d363a2a1b63b59db71a4..b5aaccc53914915535aa187390ce52950f64d1f2 100644
--- a/programs/tasklist/tasklist.rc
+++ b/programs/tasklist/tasklist.rc
@@ -39,5 +39,8 @@ STRINGTABLE
 #define WINE_FILEDESCRIPTION_STR "Wine tasklist"
 #define WINE_FILENAME_STR "tasklist.exe"
 #define WINE_FILEVERSION 6,1,7601,16385
+#define WINE_FILEVERSION_STR "6.1.7601.16385"
+#define WINE_PRODUCTVERSION 6,1,7601,16385
+#define WINE_PRODUCTVERSION_STR "6.1.7601.16385"
 
 #include "wine/wine_common_ver.rc"
