diff --git a/programs/winebrowser/Makefile.in b/programs/winebrowser/Makefile.in
index 4aaf53abcb9dbb7b424fcfb0027fe5db227ba786..328f75a2ce62f140f8e7a11230a01b10deb8ed1d 100644
--- a/programs/winebrowser/Makefile.in
+++ b/programs/winebrowser/Makefile.in
@@ -1,5 +1,5 @@
 MODULE    = winebrowser.exe
-IMPORTS   = urlmon oleaut32 shell32 user32 advapi32 ws2_32
+IMPORTS   = urlmon oleaut32 shell32 user32 advapi32
 
 EXTRADLLFLAGS = -mwindows -municode
 
