diff --git a/programs/winebrowser/main.c b/programs/winebrowser/main.c
index cb70e9ee2b7794fb7b524c9cc7a213878ecc25e6..738930e39a66fbe0633c75ccaf75514500bf9f5e 100644
--- a/programs/winebrowser/main.c
+++ b/programs/winebrowser/main.c
@@ -49,102 +49,11 @@
 #include <shellapi.h>
 #include <urlmon.h>
 #include <ddeml.h>
-#include <winsock2.h>
 
 #include "wine/debug.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(winebrowser);
 
-typedef struct {
-    int request_code;
-    char *data;
-} android_message_t;
-
-static int send_android_message(android_message_t message) {
-    WSADATA data;
-    int ret;
-
-    if ((ret = WSAStartup(MAKEWORD(2,2), &data)) != 0) {
-        TRACE( "WSAStartup failed with error %d\n", ret );
-        return 0;
-    };
-    SOCKET sock_fd = socket(AF_INET, SOCK_STREAM, 0);
-    if (sock_fd == INVALID_SOCKET) {
-        TRACE( "Failed to create client socket\n" );
-        return 0;
-    }
-    struct sockaddr_in addr;
-    ZeroMemory(&addr, sizeof(addr));
-    addr.sin_addr.s_addr = inet_addr("127.0.0.1");
-    addr.sin_family = AF_INET;
-    addr.sin_port = htons(20000);
-    ret = connect(sock_fd, (struct sockaddr *)&addr, sizeof(addr));
-    if (ret == SOCKET_ERROR) {
-        TRACE( "Failed to connect to server\n" );
-        return 0;
-    }
-    int net_requestcode = htonl(message.request_code);
-    int net_data_length = htonl(strlen(message.data));
-    ret = send(sock_fd, (const char *)&net_requestcode, sizeof(net_requestcode), 0);
-    if (ret == SOCKET_ERROR) {
-        TRACE( "Failed to send request code to server\n" );
-        return 0;
-    }
-    ret = send(sock_fd, (const char *)&net_data_length, sizeof(net_data_length), 0);
-    if (ret == SOCKET_ERROR) {
-        TRACE( "Failed to send data length to server\n" );
-        return 0;
-    }
-    ret = send(sock_fd, message.data, strlen(message.data), 0);
-    if (ret == SOCKET_ERROR) {
-        TRACE( "Failed to send data to server\n" );
-        return 0;
-    }
-
-    ret = closesocket(sock_fd);
-    if (ret == SOCKET_ERROR) {
-        TRACE( "Failed to close socket\n" );
-        return 0;
-    }
-
-    WSACleanup();
-
-    return 1;
-}
-
-static const char *from_unix_to_dos_path(const char *url) {
-    const char *path = url + 7;
-    char *temp = strdup(path);
-    char *token = strtok(temp, "/");
-    const char *new_url = NULL;
-
-    while (token != NULL) {
-        if (strcmp("dosdevices", token) == 0) {
-            token = strtok(NULL, "/");
-            new_url = strstr(path, token);
-            TRACE( "New URL is %s\n", new_url );
-            break;
-        }
-        token = strtok(NULL, "/");
-    }
-
-    free(temp);
-
-    return new_url;
-}
-
-static int launch_url_with_internet_explorer(const char *url) {
-    STARTUPINFOA si;
-    PROCESS_INFORMATION pi;
-    ZeroMemory(&si, sizeof(si));
-    si.cb = sizeof(si);
-    ZeroMemory(&pi, sizeof(pi));
-    char command[1024];
-    snprintf(command, sizeof(command), "iexplore.exe %s", url);
-    return CreateProcessA(NULL, command, NULL, NULL, FALSE, 0, NULL, NULL, &si, &pi);
-}
-
-
 static char *strdup_unixcp( const WCHAR *str )
 {
     char *ret;
@@ -154,6 +63,17 @@ static char *strdup_unixcp( const WCHAR *str )
     return ret;
 }
 
+static void restore_system_environment(void)
+{
+    const char* orig_ld_path = getenv("ORIG_LD_LIBRARY_PATH");
+
+    if (orig_ld_path)
+    {
+        __wine_set_unix_env("LD_LIBRARY_PATH", orig_ld_path);
+        __wine_set_unix_env("ORIG_LD_LIBRARY_PATH", NULL);
+    }
+}
+
 /* try to launch a unix app from a comma separated string of app names */
 static int launch_app( const WCHAR *candidates, const WCHAR *argv1 )
 {
@@ -163,6 +83,11 @@ static int launch_app( const WCHAR *candidates, const WCHAR *argv1 )
 
     if (!(cmdline = strdup_unixcp( argv1 ))) return 1;
 
+    /* PROTON HACK: Restore ORIG_LD_LIBRARY_PATH to LD_LIBRARY_PATH.
+     * System programs may not work correctly with our libraries, in
+     * particular gio on Ubuntu 19.04 is broken by our libgio. */
+    restore_system_environment();
+
     while (*candidates)
     {
         WCHAR **args = CommandLineToArgvW( candidates, &count );
@@ -181,12 +106,7 @@ static int launch_app( const WCHAR *candidates, const WCHAR *argv1 )
         free( argv_new );
         candidates += lstrlenW( candidates ) + 1;  /* grab the next app */
     }
-
-    if (strstr(cmdline, "file://")) {
-        cmdline = from_unix_to_dos_path(cmdline);
-     }
-
-    if (!launch_url_with_internet_explorer(cmdline)) WINE_ERR( "could not find a suitable app to open %s\n", debugstr_w( argv1 ));
+    WINE_ERR( "could not find a suitable app to open %s\n", debugstr_w( argv1 ));
 
     free( cmdline );
     return 1;
@@ -221,7 +141,6 @@ static int open_http_url( const WCHAR *url )
     WCHAR browsers[256];
     HKEY key;
     LONG r;
-    int wine_open_with_android_browser;
 
     /* @@ Wine registry key: HKCU\Software\Wine\WineBrowser */
     if  (!(r = RegOpenKeyW( HKEY_CURRENT_USER, L"Software\\Wine\\WineBrowser", &key )))
@@ -232,24 +151,6 @@ static int open_http_url( const WCHAR *url )
     if (r != ERROR_SUCCESS)
         memcpy( browsers, defaultbrowsers, sizeof(defaultbrowsers) );
 
-    wine_open_with_android_browser = getenv("WINE_OPEN_WITH_ANDROID_BROWSER") && atoi(getenv("WINE_OPEN_WITH_ANDROID_BROWSER"));
-
-    if (wine_open_with_android_browser) {
-        char *p = strdup_unixcp(url);
-           if (strstr(p, "file://")) {
-            if (launch_url_with_internet_explorer(from_unix_to_dos_path(p)))
-                   return 0;
-        }
-        else {
-            android_message_t message;
-            ZeroMemory(&message, sizeof(message));
-            message.request_code = 1;
-            message.data = p;
-               if (send_android_message(message))
-                  return 0;
-        }
-    }
-
     return launch_app( browsers, url );
 }
 
@@ -264,7 +165,6 @@ static int open_mailto_url( const WCHAR *url )
     WCHAR mailers[256];
     HKEY key;
     LONG r;
-    int wine_open_with_android_browser;
 
     /* @@ Wine registry key: HKCU\Software\Wine\WineBrowser */
     if (!(r = RegOpenKeyW( HKEY_CURRENT_USER, L"Software\\Wine\\WineBrowser", &key )))
@@ -473,6 +373,10 @@ static WCHAR *encode_unix_path(const char *src)
 
 static WCHAR *convert_file_uri(IUri *uri)
 {
+    UNICODE_STRING nt_name;
+    OBJECT_ATTRIBUTES attr;
+    NTSTATUS status;
+    ULONG size = 256;
     char *buffer;
     WCHAR *new_path = NULL;
     BSTR filename;
@@ -484,9 +388,23 @@ static WCHAR *convert_file_uri(IUri *uri)
 
     WINE_TRACE("Windows path: %s\n", wine_dbgstr_w(filename));
 
-    if (GetFileAttributesW( filename ) == INVALID_FILE_ATTRIBUTES) return NULL;
-    if (!(buffer = wine_get_unix_file_name( filename ))) return NULL;
-    new_path = encode_unix_path( buffer );
+    if (!RtlDosPathNameToNtPathName_U( filename, &nt_name, NULL, NULL )) return NULL;
+    InitializeObjectAttributes( &attr, &nt_name, 0, 0, NULL );
+    for (;;)
+    {
+        if (!(buffer = malloc( size )))
+        {
+            RtlFreeUnicodeString( &nt_name );
+            return NULL;
+        }
+        status = wine_nt_to_unix_file_name( &attr, buffer, &size, FILE_OPEN );
+        if (status != STATUS_BUFFER_TOO_SMALL) break;
+        free( buffer );
+    }
+
+    if (!status) new_path = encode_unix_path( buffer );
+    SysFreeString(filename);
+    free( buffer );
 
     WINE_TRACE("New path: %s\n", wine_dbgstr_w(new_path));
 
