diff --git a/programs/winecfg/winecfg.rc b/programs/winecfg/winecfg.rc
index 73a189aa5ede838fba39e471b4b9d01e24eb1b03..779a2f900e3e8081eda69d46b4c10f4e8fb51d23 100644
--- a/programs/winecfg/winecfg.rc
+++ b/programs/winecfg/winecfg.rc
@@ -267,18 +267,16 @@ FONT 8, "MS Shell Dlg"
 BEGIN
     GROUPBOX        "Driver diagnostics",IDC_STATIC,8,4,244,27
     LTEXT           "",IDC_AUDIO_DRIVER,18,16,230,8
-    PUSHBUTTON      "&Test Sound",IDC_AUDIO_TEST,177,13,69,14
-    GROUPBOX        "Defaults",IDC_STATIC,8,33,244,92
+    GROUPBOX        "Defaults",IDC_STATIC,8,33,244,76
     LTEXT           "Output device:",IDC_STATIC,18,45,230,8
     LTEXT           "Voice output device:",IDC_STATIC,18,61,230,8
     LTEXT           "Input device:",IDC_STATIC,18,77,230,8
     LTEXT           "Voice input device:",IDC_STATIC,18,93,230,8
-    LTEXT           "MIDI output device:",IDC_STATIC,18,110,230,8
     COMBOBOX        IDC_AUDIOOUT_DEVICE,110,43,135,60,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
     COMBOBOX        IDC_VOICEOUT_DEVICE,110,59,135,60,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
     COMBOBOX        IDC_AUDIOIN_DEVICE,110,75,135,60,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
     COMBOBOX        IDC_VOICEIN_DEVICE,110,91,135,60,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
-    COMBOBOX        IDC_MIDI_DEVICE,110,107,135,60,CBS_DROPDOWNLIST | WS_VSCROLL | WS_TABSTOP
+    PUSHBUTTON      "&Test Sound",IDC_AUDIO_TEST,8,113,69,14
     GROUPBOX        "Speaker configuration",IDC_STATIC,8,132,244,86
     CONTROL         "",IDC_LIST_AUDIO_DEVICES,"SysListView32",LVS_REPORT | LVS_AUTOARRANGE | LVS_ALIGNLEFT |
                     LVS_SINGLESEL | WS_BORDER | WS_TABSTOP | LVS_NOSORTHEADER, 15,146,230,52
