diff --git a/programs/wineconsole/wineconsole.c b/programs/wineconsole/wineconsole.c
index 372f8820812ee17978bca5550a7e5109d1bd73ba..aeeee11aa867c86c9dd9a2d7ff6f50edbf72f187 100644
--- a/programs/wineconsole/wineconsole.c
+++ b/programs/wineconsole/wineconsole.c
@@ -25,7 +25,6 @@
 #include "windef.h"
 #include "winbase.h"
 #include "wincon.h"
-#include "winternl.h"
 
 #include "wineconsole_res.h"
 
@@ -33,46 +32,6 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(console);
 
-static WCHAR *lookup_executable(WCHAR *cmdline)
-{
-    WCHAR path[MAX_PATH];
-    WCHAR *p;
-    BOOL status;
-
-    if (!(cmdline = wcsdup(cmdline))) return NULL;
-    if ((p = wcspbrk(cmdline, L" \t"))) *p = L'\0';
-    status = SearchPathW(NULL, cmdline, L".exe", ARRAY_SIZE(path), path, NULL);
-    free(cmdline);
-    return status ? wcsdup(path) : NULL;
-}
-
-static BOOL setup_target_console(WCHAR *title)
-{
-    BOOL ret;
-
-    if (!FreeConsole()) return FALSE;
-    /* Zero out std handles so that AllocConsole() sets the newly allocated handles as std,
-     * and will be inherited by child process.
-     */
-    SetStdHandle(STD_INPUT_HANDLE, NULL);
-    SetStdHandle(STD_OUTPUT_HANDLE, NULL);
-    SetStdHandle(STD_ERROR_HANDLE, NULL);
-    /* HACK: tweak process parameters to set the title to target executable
-     * so that conhost will take config from that target process (and not wineconsole)
-     */
-    if (title)
-    {
-        UNICODE_STRING old = RtlGetCurrentPeb()->ProcessParameters->WindowTitle;
-        RtlInitUnicodeString(&RtlGetCurrentPeb()->ProcessParameters->WindowTitle, title);
-        ret = AllocConsole();
-        RtlGetCurrentPeb()->ProcessParameters->WindowTitle = old;
-        free(title);
-    }
-    else
-        ret = AllocConsole();
-    return ret;
-}
-
 int WINAPI wWinMain( HINSTANCE inst, HINSTANCE prev, WCHAR *cmdline, INT show )
 {
     STARTUPINFOW startup = { sizeof(startup) };
@@ -84,14 +43,9 @@ int WINAPI wWinMain( HINSTANCE inst, HINSTANCE prev, WCHAR *cmdline, INT show )
 
     if (!*cmd) cmd = default_cmd;
 
-    if (!setup_target_console(lookup_executable(cmdline)))
-    {
-        ERR( "failed to allocate console: %lu\n", GetLastError() );
-        return 1;
-    }
-
-    if (!CreateProcessW( NULL, cmd, NULL, NULL, FALSE, 0, NULL, NULL, &startup, &info ))
+    if (!CreateProcessW( NULL, cmd, NULL, NULL, FALSE, CREATE_NEW_CONSOLE, NULL, NULL, &startup, &info ))
     {
+        HANDLE hStdInput, hStdOutput;
         WCHAR format[256], *buf;
         INPUT_RECORD ir;
         DWORD len;
@@ -99,19 +53,30 @@ int WINAPI wWinMain( HINSTANCE inst, HINSTANCE prev, WCHAR *cmdline, INT show )
         exit_code = GetLastError();
         WARN( "CreateProcess '%ls' failed: %lu\n", cmd, exit_code );
 
+        /* create a new console to display error messages in it */
+        FreeConsole(); /* make sure we're not connected to any console */
+        if (!AllocConsole())
+        {
+            ERR( "failed to allocate console: %lu\n", GetLastError() );
+            return 1;
+        }
+
+        hStdInput  = CreateFileW( L"CONIN$",  GENERIC_READ | GENERIC_WRITE, 0, NULL,
+                                  OPEN_EXISTING, 0, 0 );
+        hStdOutput = CreateFileW( L"CONOUT$", GENERIC_READ | GENERIC_WRITE, 0, NULL,
+                                  OPEN_EXISTING, 0, 0 );
+
         LoadStringW( GetModuleHandleW( NULL ), IDS_CMD_LAUNCH_FAILED, format, ARRAY_SIZE(format) );
         len = wcslen( format ) + wcslen( cmd );
         if ((buf = malloc( len * sizeof(WCHAR) )))
         {
             swprintf( buf, len, format, cmd );
-            WriteConsoleW( startup.hStdOutput, buf, wcslen(buf), &len, NULL);
-            while (ReadConsoleInputW( startup.hStdInput, &ir, 1, &len ) && ir.EventType == MOUSE_EVENT);
+            WriteConsoleW( hStdOutput, buf, wcslen(buf), &len, NULL);
+            while (ReadConsoleInputW( hStdInput, &ir, 1, &len ) && ir.EventType == MOUSE_EVENT);
         }
         return exit_code;
     }
 
-    /* detach from created console */
-    FreeConsole();
     CloseHandle( info.hThread );
     WaitForSingleObject( info.hProcess, INFINITE );
     return GetExitCodeProcess( info.hProcess, &exit_code ) ? exit_code : GetLastError();
