diff --git a/programs/winedbg/info.c b/programs/winedbg/info.c
index 727686c2fc510553dff8d624eebb357d2ba3bfb3..5c1cd100903855b45a34b94c106d8aacea61aabe 100644
--- a/programs/winedbg/info.c
+++ b/programs/winedbg/info.c
@@ -913,6 +913,8 @@ void info_win32_virtual(DWORD pid)
         }
         dbg_printf("%0*Ix %0*Ix %s %s %s\n",
                    ADDRWIDTH, (DWORD_PTR)addr, ADDRWIDTH, (DWORD_PTR)addr + mbi.RegionSize - 1, state, type, prot);
+        if (addr + mbi.RegionSize < addr) /* wrap around ? */
+            break;
         addr += mbi.RegionSize;
     }
     if (pid != dbg_curr_pid) CloseHandle(hProc);
