diff --git a/programs/winedbg/symbol.c b/programs/winedbg/symbol.c
index 96df020fba50c52c1fed70681999b4f7bf6d05f4..22c340337528f5abf11a857f53e1052a07193b09 100644
--- a/programs/winedbg/symbol.c
+++ b/programs/winedbg/symbol.c
@@ -154,7 +154,7 @@ static BOOL fill_sym_lvalue(const SYMBOL_INFO* sym, ULONG_PTR base,
         }
         else
         {
-            DWORD64* pdw = (DWORD64*)lexeme_alloc_size(sizeof(*pdw));
+            DWORD* pdw = (DWORD*)lexeme_alloc_size(sizeof(*pdw));
             init_lvalue(lvalue, FALSE, pdw);
             *pdw = sym->Value;
         }
