diff --git a/programs/winedevice/device.c b/programs/winedevice/device.c
index 8f816c2c0ecf3c2234ea5d0c2f5a097c17ca884b..bb5850872305ae6882a0b35bdbffb2f6c705eb0d 100644
--- a/programs/winedevice/device.c
+++ b/programs/winedevice/device.c
@@ -28,6 +28,7 @@
 #include "ddk/wdm.h"
 #include "wine/svcctl.h"
 #include "wine/debug.h"
+#include "wine/heap.h"
 
 WINE_DEFAULT_DEBUG_CHANNEL(ntoskrnl);
 
@@ -63,7 +64,7 @@ static DWORD device_handler( DWORD ctrl, const WCHAR *driver_name )
     DWORD result = NO_ERROR;
     WCHAR *str;
 
-    if (!(str = malloc( sizeof(servicesW) + lstrlenW(driver_name)*sizeof(WCHAR) )))
+    if (!(str = heap_alloc( sizeof(servicesW) + lstrlenW(driver_name)*sizeof(WCHAR) )))
         return STATUS_NO_MEMORY;
 
     lstrcpyW( str, servicesW );
@@ -89,7 +90,7 @@ static DWORD device_handler( DWORD ctrl, const WCHAR *driver_name )
         break;
     }
 
-    free( str );
+    RtlFreeUnicodeString( &service_name );
     return result;
 }
 
