diff --git a/programs/winetest/main.c b/programs/winetest/main.c
index fe6542d9c3249f90cb246464f922e88fe326eaf8..3ded6632cd5078876f7fea30fdaa1df1e341583b 100644
--- a/programs/winetest/main.c
+++ b/programs/winetest/main.c
@@ -68,7 +68,6 @@ static char build_id[64];
 static BOOL is_wow64;
 static int failures;
 static int quiet_mode;
-static int test_timeout = 120;
 static HANDLE logfile;
 static HANDLE junit;
 
@@ -472,8 +471,6 @@ static void print_version (void)
     void (CDECL *wine_get_host_version)( const char **sysname, const char **release );
     BOOL (WINAPI *pGetProductInfo)(DWORD, DWORD, DWORD, DWORD, DWORD *);
     NTSTATUS (WINAPI *pRtlGetVersion)(RTL_OSVERSIONINFOEXW *);
-    DWORD revision, size = sizeof(revision);
-    HKEY hkey;
 
     ver.dwOSVersionInfoSize = sizeof(ver);
     if (!(ext = GetVersionExA ((OSVERSIONINFOA *) &ver)))
@@ -526,13 +523,6 @@ static void print_version (void)
              ver.dwMajorVersion, ver.dwMinorVersion, ver.dwBuildNumber,
              ver.dwPlatformId, ver.szCSDVersion);
 
-    if (!RegOpenKeyA( HKEY_LOCAL_MACHINE, "Software\\Microsoft\\Windows NT\\CurrentVersion", &hkey ))
-    {
-        if (!RegQueryValueExA( hkey, "UBR", NULL, NULL, (BYTE *)&revision, &size ))
-            xprintf( "    UBR=%lu\n", revision );
-        RegCloseKey( hkey );
-    }
-
     wine_get_build_id = (void *)GetProcAddress(hntdll, "wine_get_build_id");
     wine_get_host_version = (void *)GetProcAddress(hntdll, "wine_get_host_version");
     if (wine_get_build_id) xprintf( "    WineBuild=%s\n", wine_get_build_id() );
@@ -658,8 +648,7 @@ static void* extract_rcdata (LPCSTR name, LPCSTR type, DWORD* size)
     HRSRC rsrc;
     HGLOBAL hdl;
     LPVOID addr;
-
-    *size = 0;
+    
     if (!(rsrc = FindResourceA(NULL, name, type)) ||
         !(*size = SizeofResource (0, rsrc)) ||
         !(hdl = LoadResource (0, rsrc)) ||
@@ -914,45 +903,36 @@ static void report_test_start( struct wine_test *test, const char *subtest, cons
 }
 
 /* filter out color escapes and null characters from test output data */
-static void *filter_data( const char *data, DWORD size, DWORD *output_size )
+static void *filter_data( const char *data, DWORD size )
 {
-    DWORD i, j, eol, ignore = 0;
-    char *ret;
+    DWORD i, j;
+    char *ret = malloc( size + 1 );
 
-    if (!(ret = malloc( size + 1 ))) return NULL;
-    for (i = j = 0, eol = -1; i < size; i++)
+    for (i = j = 0; i < size; i++)
     {
         if (data[i] == '\x1b' && data[i + 1] == '[')
         {
             while (data[i] && data[i] != 'm') i++;
-            eol = i;
         }
         else if (data[i]) ret[j++] = data[i];
-        if (!strncmp( data + i, " Test succeeded", 15 )) ignore += i + 15 - eol;
-        if (data[i] == '\n') eol = i;
     }
     ret[j] = 0;
-    *output_size = j - ignore;
-
     return ret;
 }
 
 static void report_test_done( struct wine_test *test, const char *subtest, const char *file, DWORD pid, DWORD ticks,
-                              HANDLE out_file, UINT status, const char *data, DWORD size, DWORD *output_size )
+                              HANDLE out_file, UINT status, const char *data, DWORD size )
 {
-    char *filtered_data;
-
-    if (!(filtered_data = filter_data( data, size, output_size ))) return;
-
-    if (quiet_mode <= 1 || status || *output_size > MAX_OUTPUT_SIZE) WriteFile( out_file, data, size, &size, NULL );
+    if (quiet_mode <= 1 || status || size > MAX_OUTPUT_SIZE) WriteFile( out_file, data, size, &size, NULL );
     xprintf( "%s:%s:%04lx done (%d) in %lds %luB\n", test->name, subtest, pid, status, ticks / 1000, size );
-    if (*output_size > MAX_OUTPUT_SIZE) xprintf( "%s:%s:%04lx The test prints too much data (%lu bytes)\n", test->name, subtest, pid, size );
+    if (size > MAX_OUTPUT_SIZE) xprintf( "%s:%s:%04lx The test prints too much data (%lu bytes)\n", test->name, subtest, pid, size );
 
-    if (filtered_data && junit)
+    if (junit)
     {
         int total = 0, fail_total = 0, skip_total = 0, failures = 0;
         const char *next, *line, *ptr, *dir = strrchr( file, '/' );
         float time, last_time = 0;
+        char *filtered_data = filter_data( data, size );
 
         for (line = next = filtered_data; *line; line = next)
         {
@@ -1038,7 +1018,7 @@ static void report_test_done( struct wine_test *test, const char *subtest, const
             output( junit, "<system-out>Test exited with status %d</system-out><failure/>", status );
             output( junit, "</testcase>\n" );
         }
-        if (*output_size > MAX_OUTPUT_SIZE)
+        if (size > MAX_OUTPUT_SIZE)
         {
             output( junit, "    <testcase classname=\"%s:%s\" name=\"%s:%s output overflow\" file=\"%s\" assertions=\"%d\" time=\"%f\">",
                      test->name, subtest, test->name, subtest, file, total, ticks / 1000.0 );
@@ -1048,9 +1028,8 @@ static void report_test_done( struct wine_test *test, const char *subtest, const
         }
 
         output( junit, "  </testsuite>\n" );
+        free( filtered_data );
     }
-
-    free( filtered_data );
 }
 
 static void report_test_skip( struct wine_test *test, const char *subtest, const char *file )
@@ -1075,22 +1054,22 @@ run_test (struct wine_test* test, const char* subtest, HANDLE out_file, const ch
         char *data, tmpname[MAX_PATH];
         HANDLE tmpfile = create_temp_file( tmpname );
         int status;
-        DWORD pid, size, output_size = 0, start = GetTickCount();
+        DWORD pid, size, start = GetTickCount();
         char *cmd = strmake("%s %s", test->exename, subtest);
 
         report_test_start( test, subtest, file );
         /* Flush to disk so we know which test caused Windows to crash if it does */
         FlushFileBuffers(out_file);
 
-        status = run_ex( cmd, tmpfile, tempdir, test_timeout * 1000, FALSE, &pid );
+        status = run_ex( cmd, tmpfile, tempdir, 120000, FALSE, &pid );
         if (status == -2 && GetLastError()) status = -GetLastError();
         free(cmd);
 
         data = flush_temp_file( tmpname, tmpfile, &size );
-        report_test_done( test, subtest, file, pid, GetTickCount() - start, out_file, status, data, size, &output_size );
+        report_test_done( test, subtest, file, pid, GetTickCount() - start, out_file, status, data, size );
         free( data );
 
-        if (status || output_size > MAX_OUTPUT_SIZE) failures++;
+        if (status || size > MAX_OUTPUT_SIZE) failures++;
     }
     if (failures) report (R_STATUS, "Running tests - %u failures", failures);
 }
@@ -1174,23 +1153,6 @@ static void get_dll_path(HMODULE dll, char **path, char *filename)
     *path = xstrdup( dllpath );
 }
 
-static const char *get_compiler_version(void)
-{
-#ifdef __clang__
-# ifdef _MSC_VER
-    return strmake( "clang %s (msvc %u)", __clang_version__, _MSC_VER );
-# else
-    return strmake( "clang %s", __clang_version__ );
-# endif
-#elif defined __GNUC__
-    return strmake( "gcc %u.%u.%u (%s)", __GNUC__, __GNUC_MINOR__, __GNUC_PATCHLEVEL__, __VERSION__ );
-#elif defined _MSC_VER
-    return strmake( "msvc %u", _MSC_VER );
-#else
-    return "unknown";
-#endif
-}
-
 static BOOL CALLBACK
 extract_test_proc (HMODULE hModule, LPCSTR lpszType, LPSTR lpszName, LONG_PTR lParam)
 {
@@ -1313,6 +1275,8 @@ static char *
 run_tests (char *logname, char *outdir)
 {
     int i;
+    char *strres, *eol, *nextline;
+    DWORD strsize;
     char tmppath[MAX_PATH], tempdir[MAX_PATH+4];
     BOOL newdir;
     DWORD needed;
@@ -1374,9 +1338,23 @@ run_tests (char *logname, char *outdir)
     }
     xprintf ("Version 4\n");
     xprintf ("Tests from build %s\n", build_id[0] ? build_id : "-" );
+    xprintf ("Archive: -\n");  /* no longer used */
     xprintf ("Tag: %s\n", tag);
     xprintf ("Build info:\n");
-    xprintf ("    Compiler: %s\n", get_compiler_version());
+    strres = extract_rcdata ("BUILD_INFO", "STRINGRES", &strsize);
+    while (strres) {
+        eol = memchr (strres, '\n', strsize);
+        if (!eol) {
+            nextline = NULL;
+            eol = strres + strsize;
+        } else {
+            strsize -= eol - strres + 1;
+            nextline = strsize?eol+1:NULL;
+            if (eol > strres && *(eol-1) == '\r') eol--;
+        }
+        xprintf ("    %.*s\n", (int)(eol-strres), strres);
+        strres = nextline;
+    }
     xprintf ("Operating system version:\n");
     print_version ();
     print_language ();
@@ -1534,7 +1512,6 @@ usage (void)
 " -S URL    URL to submit the results to\n"
 " -t TAG    include TAG of characters [-.0-9a-zA-Z] in the report\n"
 " -u URL    include TestBot URL in the report\n"
-" -w SECS   how many seconds to wait for each test to finish (default: 120)\n"
 " -x DIR    Extract tests to DIR (default: .\\wct) and exit\n");
 }
 
@@ -1595,13 +1572,6 @@ int __cdecl main( int argc, char *argv[] )
                 exit( 2 );
             }
             break;
-        case 'w':
-            if (!argv[++i] || !(test_timeout = atoi( argv[i] )))
-            {
-                usage();
-                exit( 2 );
-            }
-            break;
         case 'J':
             if (!(path = argv[++i]))
             {
@@ -1719,7 +1689,6 @@ int __cdecl main( int argc, char *argv[] )
             SetEnvironmentVariableA( "WINETEST_PLATFORM", running_under_wine () ? "wine" : "windows" );
             SetEnvironmentVariableA( "WINETEST_DEBUG", "1" );
             SetEnvironmentVariableA( "WINETEST_INTERACTIVE", "0" );
-            SetEnvironmentVariableA( "WINETEST_MUTE_THRESHOLD", "4" );
             SetEnvironmentVariableA( "WINETEST_REPORT_SUCCESS", "0" );
         }
         if (junit)
