diff --git a/programs/winver/version.rc b/programs/winver/version.rc
index 77d8314a4732e8fa1051f1914391bb164303d05f..c079dde644a3af6fcd70e32808003b51e90e6fe0 100644
--- a/programs/winver/version.rc
+++ b/programs/winver/version.rc
@@ -22,6 +22,9 @@
 #define WINE_FILENAME_STR "winver.exe"
 #define WINE_FILETYPE VFT_APP
 #define WINE_FILEVERSION 5,1,2600,0
+#define WINE_FILEVERSION_STR "5.1.2600.0"
+#define WINE_PRODUCTVERSION 5,1,2600,0
+#define WINE_PRODUCTVERSION_STR "5.1.2600.0"
 
 #include "wine/wine_common_ver.rc"
 
