diff --git a/programs/wmic/main.c b/programs/wmic/main.c
index 9d1748ca7533850b3c80c4a8c730f5b2eeb8c1af..6ca3e680654965e31f3b70a157420bf4ada4598e 100644
--- a/programs/wmic/main.c
+++ b/programs/wmic/main.c
@@ -55,7 +55,6 @@ alias_map[] =
     { L"nicconfig", L"Win32_NetworkAdapterConfiguration" },
     { L"os", L"Win32_OperatingSystem" },
     { L"process", L"Win32_Process" },
-    { L"qfe", L"Win32_QuickFixEngineering" },
     { L"systemenclosure", L"Win32_SystemEnclosure" },
 };
 
@@ -179,9 +178,6 @@ static void convert_to_bstr( VARIANT *v )
     BSTR out = NULL;
     VARTYPE vt;
 
-    /* treat VT_I4 values as unsigned */
-    if (V_VT(v) == VT_I4) V_VT(v) = VT_UI4;
-
     if (SUCCEEDED(VariantChangeType( v, v, 0, VT_BSTR ))) return;
     vt = V_VT(v);
     if (vt == (VT_ARRAY | VT_BSTR))
@@ -354,12 +350,6 @@ static int process_args( int argc, WCHAR *argv[] )
     const WCHAR *class;
     int i;
 
-    if (!wcscmp( argv[0], L"/?" ))
-    {
-        output_error( STRING_HELP );
-        return 0;
-    }
-
     for (i = 0; i < argc && argv[i][0] == '/'; i++)
         WINE_FIXME( "switch %s not supported\n", debugstr_w(argv[i]) );
 
