diff --git a/programs/wmic/wmic.rc b/programs/wmic/wmic.rc
index 431bbf6e4747d0b8481402f1a108f6535d4bde26..80d76ac1447d9d41a6617930f5904bd9b8ea0f93 100644
--- a/programs/wmic/wmic.rc
+++ b/programs/wmic/wmic.rc
@@ -29,5 +29,4 @@ STRINGTABLE
     STRING_INVALID_QUERY, "Error: Invalid query\n"
     STRING_INVALID_PATH, "Error: Invalid syntax for PATH\n"
     STRING_USAGE, "Supply a command\n"
-    STRING_HELP, "Example usage: wmic ALIAS get PROPERTY[,PROPERTY2,...]\n"
 }
