diff --git a/programs/wuauserv/main.c b/programs/wuauserv/main.c
index b3e68f580b65e0b073c35755ec038ae3acb6b028..79eb84e67030d14b9e52101eeb328407ffdaae2a 100644
--- a/programs/wuauserv/main.c
+++ b/programs/wuauserv/main.c
@@ -26,7 +26,7 @@
 
 WINE_DEFAULT_DEBUG_CHANNEL(wuauserv);
 
-static WCHAR wuauservW[] = L"wuauserv";
+static WCHAR wuauservW[] = {'w','u','a','u','s','e','r','v',0};
 
 static SERVICE_STATUS_HANDLE service_handle;
 static HANDLE stop_event;
