diff --git a/programs/xcopy/xcopy.rc b/programs/xcopy/xcopy.rc
index ae276c6b876875c4c07ba124b882f9e11a0b3f5e..c61e4da88fd2b50317b339b09751d6fba80d26ef 100644
--- a/programs/xcopy/xcopy.rc
+++ b/programs/xcopy/xcopy.rc
@@ -79,5 +79,4 @@ Where:\n\
 \t\tIf no date is supplied, only copy if destination is older\n\
 \t\tthan source.\n\n"
 
-    STRING_NOCOPYTOSELF, "File cannot be copied onto itself.\n"
 }
