diff --git a/server/clipboard.c b/server/clipboard.c
index 59a50354b48ea0eb808ac1e7173e6776f6aabd31..a2e0af2ba32fcc4396a34d40afae6ffb76233681 100644
--- a/server/clipboard.c
+++ b/server/clipboard.c
@@ -76,10 +76,11 @@ static const struct object_ops clipboard_ops =
     no_add_queue,                 /* add_queue */
     NULL,                         /* remove_queue */
     NULL,                         /* signaled */
+    NULL,                         /* get_esync_fd */
+    NULL,                         /* get_fsync_idx */
     NULL,                         /* satisfied */
     no_signal,                    /* signal */
     no_get_fd,                    /* get_fd */
-    default_get_sync,             /* get_sync */
     default_map_access,           /* map_access */
     default_get_sd,               /* get_sd */
     default_set_sd,               /* set_sd */
