diff --git a/server/handle.c b/server/handle.c
index ddcf03accb2d81b065a453e8ddfeb674b4398543..2dca6d8155d7eb12f76a8651b97b9e2d2566479c 100644
--- a/server/handle.c
+++ b/server/handle.c
@@ -126,10 +126,11 @@ static const struct object_ops handle_table_ops =
     no_add_queue,                    /* add_queue */
     NULL,                            /* remove_queue */
     NULL,                            /* signaled */
+    NULL,                            /* get_esync_fd */
+    NULL,                            /* get_fsync_idx */
     NULL,                            /* satisfied */
     no_signal,                       /* signal */
     no_get_fd,                       /* get_fd */
-    default_get_sync,                /* get_sync */
     default_map_access,              /* map_access */
     default_get_sd,                  /* get_sd */
     default_set_sd,                  /* set_sd */
@@ -284,16 +285,20 @@ obj_handle_t alloc_handle_no_access_check( struct process *process, void *ptr, u
 obj_handle_t alloc_handle( struct process *process, void *ptr, unsigned int access, unsigned int attr )
 {
     struct object *obj = ptr;
-
-    if (!(access = obj->ops->map_access( obj, access ) & ~RESERVED_ALL))
-    {
-        set_error( STATUS_ACCESS_DENIED );
-        return 0;
-    }
-    if (!check_object_access( NULL, obj, &access )) return 0;
+    access = obj->ops->map_access( obj, access ) & ~RESERVED_ALL;
+    if (access && !check_object_access( NULL, obj, &access )) return 0;
     return alloc_handle_entry( process, ptr, access, attr );
 }
 
+/* allocate handle for opening an object by userspace request */
+obj_handle_t alloc_handle_user_open( struct process *process, void *obj, unsigned int access, unsigned int attr )
+{
+    if (access) return alloc_handle( process, obj, access, attr );
+
+    set_error( STATUS_ACCESS_DENIED );
+    return 0;
+}
+
 /* allocate a global handle for an object, incrementing its refcount */
 /* return the handle, or 0 on error */
 static obj_handle_t alloc_global_handle_no_access_check( void *obj, unsigned int access )
@@ -429,17 +434,15 @@ unsigned int close_handle( struct process *process, obj_handle_t handle )
     struct handle_table *table;
     struct handle_entry *entry;
     struct object *obj;
-    int index = handle_to_index( handle );
 
     if (!(entry = get_handle( process, handle ))) return STATUS_INVALID_HANDLE;
     if (entry->access & RESERVED_CLOSE_PROTECT) return STATUS_HANDLE_NOT_CLOSABLE;
     obj = entry->ptr;
     if (!obj->ops->close_handle( obj, process, handle )) return STATUS_HANDLE_NOT_CLOSABLE;
-
+    entry->ptr = NULL;
     table = handle_is_global(handle) ? global_table : process->handles;
-    table->entries[index].ptr = NULL;
-    if (index < table->free) table->free = index;
-    if (index == table->last) shrink_handle_table( table );
+    if (entry < table->entries + table->free) table->free = entry - table->entries;
+    if (entry == table->entries + table->last) shrink_handle_table( table );
     release_object_from_handle( obj );
     return STATUS_SUCCESS;
 }
@@ -653,7 +656,7 @@ obj_handle_t open_object( struct process *process, obj_handle_t parent, unsigned
 
     if ((obj = open_named_object( root, ops, name, attributes )))
     {
-        handle = alloc_handle( process, obj, access, attributes );
+        handle = alloc_handle_user_open( process, obj, access, attributes );
         release_object( obj );
     }
     if (root) release_object( root );
@@ -725,7 +728,7 @@ DECL_HANDLER(get_object_name)
 
     if (!(obj = get_handle_obj( current->process, req->handle, 0, NULL ))) return;
 
-    if ((name = obj->ops->get_full_name( obj, get_reply_max_size(), &reply->total )))
+    if ((name = obj->ops->get_full_name( obj, &reply->total )))
         set_reply_data_ptr( name, min( reply->total, get_reply_max_size() ));
     release_object( obj );
 }
@@ -864,6 +867,7 @@ static int enum_handles( struct process *process, void *user )
         handle->access     = entry->access & ~RESERVED_ALL;
         handle->type       = entry->ptr->ops->type->index;
         handle->attributes = 0;
+        handle->object     = (client_ptr_t)(ULONG_PTR)entry->ptr | (client_ptr_t)0xffff800000000000;
         if (entry->access & RESERVED_INHERIT) handle->attributes |= OBJ_INHERIT;
         if (entry->access & RESERVED_CLOSE_PROTECT) handle->attributes |= OBJ_PROTECT_CLOSE;
         info->count--;
