diff --git a/server/hook.c b/server/hook.c
index fe3c0459651853b061b8cabaa8ab3f793fc4f18b..499b333dc3a2fb8e46283c5285cb7f695afaa7d6 100644
--- a/server/hook.c
+++ b/server/hook.c
@@ -58,6 +58,11 @@ struct hook
     data_size_t         module_size;
 };
 
+#define WH_WINEVENT (WH_MAXHOOK+1)
+
+#define NB_HOOKS (WH_WINEVENT-WH_MINHOOK+1)
+#define HOOK_ENTRY(p)  LIST_ENTRY( (p), struct hook, chain )
+
 struct hook_table
 {
     struct object obj;              /* object header */
@@ -76,10 +81,11 @@ static const struct object_ops hook_table_ops =
     no_add_queue,                 /* add_queue */
     NULL,                         /* remove_queue */
     NULL,                         /* signaled */
+    NULL,                         /* get_esync_fd */
+    NULL,                         /* get_fsync_idx */
     NULL,                         /* satisfied */
     no_signal,                    /* signal */
     no_get_fd,                    /* get_fd */
-    default_get_sync,             /* get_sync */
     default_map_access,           /* map_access */
     default_get_sd,               /* get_sd */
     default_set_sd,               /* set_sd */
@@ -149,7 +155,7 @@ static struct hook *add_hook( struct desktop *desktop, struct process *process,
     }
     if (!(hook = mem_alloc( sizeof(*hook) ))) return NULL;
 
-    if (!(hook->handle = alloc_user_handle( hook, NULL, NTUSER_OBJ_HOOK )))
+    if (!(hook->handle = alloc_user_handle( hook, USER_HOOK )))
     {
         free( hook );
         return NULL;
@@ -202,16 +208,27 @@ static void free_hook( struct hook *hook )
 /* find a hook from its index and proc */
 static struct hook *find_hook( struct thread *thread, int index, client_ptr_t proc )
 {
-    struct hook *hook;
+    struct list *p;
     struct hook_table *table = get_queue_hooks( thread );
 
     if (table)
-        LIST_FOR_EACH_ENTRY( hook, &table->hooks[index], struct hook, chain )
+    {
+        LIST_FOR_EACH( p, &table->hooks[index] )
+        {
+            struct hook *hook = HOOK_ENTRY( p );
             if (hook->proc == proc) return hook;
-
+        }
+    }
     return NULL;
 }
 
+/* get the first hook in the chain */
+static inline struct hook *get_first_hook( struct hook_table *table, int index )
+{
+    struct list *elem = list_head( &table->hooks[index] );
+    return elem ? HOOK_ENTRY( elem ) : NULL;
+}
+
 /* check if a given hook should run in the owner thread instead of the current thread */
 static inline int run_hook_in_owner_thread( struct hook *hook )
 {
@@ -235,9 +252,9 @@ static inline struct hook *get_first_valid_hook( struct hook_table *table, int i
                                                  int event, user_handle_t win,
                                                  int object_id, int child_id )
 {
-    struct hook *hook;
+    struct hook *hook = get_first_hook( table, index );
 
-    LIST_FOR_EACH_ENTRY( hook, &table->hooks[index], struct hook, chain )
+    while (hook)
     {
         if (hook->proc && run_hook_in_current_thread( hook ))
         {
@@ -252,8 +269,9 @@ static inline struct hook *get_first_valid_hook( struct hook_table *table, int i
                                 hook->handle );
             }
         }
+        hook = HOOK_ENTRY( list_next( &table->hooks[index], &hook->chain ) );
     }
-    return NULL;
+    return hook;
 }
 
 /* find the next hook in the chain, skipping the deleted ones */
@@ -261,12 +279,10 @@ static struct hook *get_next_hook( struct thread *thread, struct hook *hook, int
                                    user_handle_t win, int object_id, int child_id )
 {
     struct hook_table *global_hooks, *table = hook->table;
-    struct list *ptr;
     int index = hook->index;
 
-    while ((ptr = list_next( &table->hooks[index], &hook->chain )))
+    while ((hook = HOOK_ENTRY( list_next( &table->hooks[index], &hook->chain ) )))
     {
-        hook = LIST_ENTRY( ptr, struct hook, chain );
         if (hook->proc && run_hook_in_current_thread( hook ))
         {
             if (event >= hook->event_min && event <= hook->event_max)
@@ -282,9 +298,11 @@ static struct hook *get_next_hook( struct thread *thread, struct hook *hook, int
         }
     }
     global_hooks = get_global_hooks( thread );
-    if (!global_hooks || global_hooks == table) return NULL;
-    /* now search through the global table */
-    return get_first_valid_hook( global_hooks, index, event, win, object_id, child_id );
+    if (global_hooks && table != global_hooks)  /* now search through the global table */
+    {
+        hook = get_first_valid_hook( global_hooks, index, event, win, object_id, child_id );
+    }
+    return hook;
 }
 
 static void hook_table_dump( struct object *obj, int verbose )
@@ -296,12 +314,13 @@ static void hook_table_dump( struct object *obj, int verbose )
 static void hook_table_destroy( struct object *obj )
 {
     int i;
-    struct hook *hook, *next;
+    struct hook *hook;
     struct hook_table *table = (struct hook_table *)obj;
 
     for (i = 0; i < NB_HOOKS; i++)
-        LIST_FOR_EACH_ENTRY_SAFE( hook, next, &table->hooks[i], struct hook, chain )
-            free_hook( hook );
+    {
+        while ((hook = get_first_hook( table, i )) != NULL) free_hook( hook );
+    }
 }
 
 /* remove a hook, freeing it if the chain is not in use */
@@ -354,25 +373,34 @@ static void release_hook_chain( struct hook_table *table, int index )
     }
     if (!--table->counts[index])
     {
-        struct hook *hook, *next;
-
-        LIST_FOR_EACH_ENTRY_SAFE( hook, next, &table->hooks[index], struct hook, chain )
+        struct hook *hook = get_first_hook( table, index );
+        while (hook)
+        {
+            struct hook *next = HOOK_ENTRY( list_next( &table->hooks[hook->index], &hook->chain ) );
             if (!hook->proc) free_hook( hook );
+            hook = next;
+        }
     }
 }
 
 /* remove all global hooks owned by a given thread */
 void remove_thread_hooks( struct thread *thread )
 {
-    struct hook *hook, *next;
     struct hook_table *global_hooks = get_global_hooks( thread );
     int index;
 
     if (!global_hooks) return;
 
     for (index = 0; index < NB_HOOKS; index++)
-        LIST_FOR_EACH_ENTRY_SAFE( hook, next, &global_hooks->hooks[index], struct hook, chain )
+    {
+        struct hook *hook = get_first_hook( global_hooks, index );
+        while (hook)
+        {
+            struct hook *next = HOOK_ENTRY( list_next( &global_hooks->hooks[index], &hook->chain ) );
             if (hook->thread == thread || hook->owner == thread) remove_hook( hook );
+            hook = next;
+        }
+    }
 }
 
 /* return the thread that owns the first global hook */
@@ -476,7 +504,7 @@ DECL_HANDLER(remove_hook)
 
     if (req->handle)
     {
-        if (!(hook = get_user_object( req->handle, NTUSER_OBJ_HOOK )))
+        if (!(hook = get_user_object( req->handle, USER_HOOK )))
         {
             set_error( STATUS_INVALID_HANDLE );
             return;
@@ -563,7 +591,7 @@ DECL_HANDLER(get_hook_info)
 {
     struct hook *hook;
 
-    if (!(hook = get_user_object( req->handle, NTUSER_OBJ_HOOK ))) return;
+    if (!(hook = get_user_object( req->handle, USER_HOOK ))) return;
     if (hook->thread && (hook->thread != current))
     {
         set_error( STATUS_INVALID_HANDLE );
