diff --git a/server/mach.c b/server/mach.c
index 0f75d892816002d73b0c19796ed71c2620e74b02..c3d4a33bdc24a6160d0afab8047d8859fc7c975b 100644
--- a/server/mach.c
+++ b/server/mach.c
@@ -155,9 +155,6 @@ void init_process_tracing( struct process *process )
                 mach_port_deallocate( mach_task_self(), msg.task_port.name );
         }
     }
-    /* On Mach thread priorities depend on having the process port available, so
-     * reapply all thread priorities here after process tracing is initialized */
-    set_process_base_priority( process, process->base_priority );
 }
 
 /* terminate the per-process tracing mechanism */
@@ -415,15 +412,13 @@ int read_process_memory( struct process *process, client_ptr_t ptr, data_size_t
 }
 
 /* write data to a process memory space */
-int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, const char *src,
-                          data_size_t *written )
+int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, const char *src )
 {
     kern_return_t ret;
+    unsigned int page_size = get_page_size();
     mach_port_t process_port = get_process_port( process );
     mach_vm_offset_t data;
 
-    if (written) *written = 0;
-
     if (!process_port)
     {
         set_error( STATUS_ACCESS_DENIED );
@@ -434,7 +429,7 @@ int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t
         set_error( STATUS_ACCESS_DENIED );
         return 0;
     }
-    if (posix_memalign( (void **)&data, get_page_size(), size ))
+    if (posix_memalign( (void **)&data, page_size, size ))
     {
         set_error( STATUS_NO_MEMORY );
         return 0;
@@ -455,7 +450,7 @@ int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t
         mach_vm_address_t current_address = (mach_vm_address_t)ptr;
         mach_vm_address_t region_address = current_address;
         mach_vm_size_t region_size, write_size;
-        vm_region_basic_info_data_64_t info;
+        vm_region_basic_info_data_t info;
         mach_msg_type_number_t info_count = VM_REGION_BASIC_INFO_COUNT_64;
         mach_port_t object_name;
         data_size_t remaining_size = size;
@@ -531,9 +526,8 @@ int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t
                     info.protection );
             if (ret != KERN_SUCCESS) break;
 
-            if (written) *written += write_size;
-            current_address       += write_size;
-            remaining_size        -= write_size;
+            current_address += write_size;
+            remaining_size  -= write_size;
         }
 
         task_resume( process_port );
@@ -542,8 +536,48 @@ int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t
 out:
     free( (void *)data );
     mach_set_error( ret );
-    if (ret == KERN_SUCCESS && written) *written = size;
     return (ret == KERN_SUCCESS);
 }
 
+/* retrieve an LDT selector entry */
+void get_selector_entry( struct thread *thread, int entry, unsigned int *base,
+                         unsigned int *limit, unsigned char *flags )
+{
+    const unsigned int total_size = (2 * sizeof(int) + 1) * 8192;
+    struct process *process = thread->process;
+    mach_vm_address_t data;
+    kern_return_t ret;
+    mach_vm_size_t bytes_read;
+    mach_port_t process_port = get_process_port( thread->process );
+
+    if (!process->ldt_copy || !process_port)
+    {
+        set_error( STATUS_ACCESS_DENIED );
+        return;
+    }
+    if (entry >= 8192)
+    {
+        set_error( STATUS_ACCESS_VIOLATION );
+        return;
+    }
+
+    if (!(data = (mach_vm_address_t)malloc( total_size )))
+    {
+        set_error( STATUS_NO_MEMORY );
+        return;
+    }
+
+    ret = mach_vm_read_overwrite( process_port, (mach_vm_address_t)process->ldt_copy, (mach_vm_size_t)total_size, data, &bytes_read );
+    if (ret != KERN_SUCCESS) mach_set_error( ret );
+    else
+    {
+        const int *ldt = (const int *)data;
+        memcpy( base, ldt + entry, sizeof(int) );
+        memcpy( limit, ldt + entry + 8192, sizeof(int) );
+        memcpy( flags, (char *)(ldt + 2 * 8192) + entry, 1 );
+    }
+
+    free( (void *)data );
+}
+
 #endif  /* USE_MACH */
