diff --git a/server/procfs.c b/server/procfs.c
index 4fb3136e080b4ee9c1721d86d4ff0edab5463aa1..6ae0a9006ed5c65b06ee14e882bc5814ca5f85e7 100644
--- a/server/procfs.c
+++ b/server/procfs.c
@@ -150,8 +150,7 @@ int read_process_memory( struct process *process, client_ptr_t ptr, size_t size,
 }
 
 /* write data to a process memory space */
-int write_process_memory( struct process *process, client_ptr_t ptr, size_t size, const char *src,
-                          data_size_t *written )
+int write_process_memory( struct process *process, client_ptr_t ptr, size_t size, const char *src )
 {
     ssize_t ret;
     int fd;
@@ -166,7 +165,6 @@ int write_process_memory( struct process *process, client_ptr_t ptr, size_t size
 
     ret = pwrite( fd, src, size, (off_t)ptr );
     close( fd );
-    if (ret == size && written) *written = size;
     if (ret == size) return 1;
 
     if (ret == -1) file_set_error();
@@ -174,6 +172,36 @@ int write_process_memory( struct process *process, client_ptr_t ptr, size_t size
     return 0;
 }
 
+/* retrieve an LDT selector entry */
+void get_selector_entry( struct thread *thread, int entry, unsigned int *base,
+                         unsigned int *limit, unsigned char *flags )
+{
+    ssize_t ret;
+    off_t pos = thread->process->ldt_copy;
+    int fd;
+
+    if (!pos)
+    {
+        set_error( STATUS_ACCESS_DENIED );
+        return;
+    }
+    if ((fd = open_proc_as( thread->process, O_RDONLY )) == -1) return;
+
+    ret = pread( fd, base, sizeof(*base), pos + entry*sizeof(int) );
+    if (ret != sizeof(*base)) goto error;
+    ret = pread( fd, limit, sizeof(*limit), pos + (8192 + entry)*sizeof(int) );
+    if (ret != sizeof(*limit)) goto error;
+    ret = pread( fd, flags, sizeof(*flags), pos + 2*8192*sizeof(int) + entry );
+    if (ret != sizeof(*flags)) goto error;
+    close( fd );
+    return;
+
+error:
+    if (ret == -1) file_set_error();
+    else set_error( STATUS_ACCESS_VIOLATION );
+    close( fd );
+}
+
 /* initialize registers in new thread if necessary */
 void init_thread_context( struct thread *thread )
 {
