diff --git a/server/ptrace.c b/server/ptrace.c
index c4b71e99de3e7680fd6c711ada4a85b3d7bfb4be..a40c5f40847ec052f04270ee8bf9741f65894453 100644
--- a/server/ptrace.c
+++ b/server/ptrace.c
@@ -43,11 +43,10 @@
 #ifdef HAVE_SYS_THR_H
 # include <sys/thr.h>
 #endif
-#ifdef HAVE_SYS_UIO_H
-# include <sys/uio.h>
-#endif
 #include <unistd.h>
 
+#include <sys/uio.h>
+
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
 #include "winternl.h"
@@ -182,11 +181,18 @@ void sigchld_callback(void)
 static int get_ptrace_pid( struct thread *thread )
 {
 #ifdef linux  /* linux always uses thread id */
-    return thread->unix_tid;
+    if (thread->unix_tid != -1) return thread->unix_tid;
 #endif
     return thread->unix_pid;
 }
 
+/* return the Unix tid to use in ptrace calls for a given thread */
+static int get_ptrace_tid( struct thread *thread )
+{
+    if (thread->unix_tid != -1) return thread->unix_tid;
+    return thread->unix_pid;
+}
+
 /* wait for a ptraced child to get a certain signal */
 static int waitpid_thread( struct thread *thread, int signal )
 {
@@ -255,17 +261,21 @@ int send_thread_signal( struct thread *thread, int sig )
 {
     int ret = -1;
 
-    if (thread->unix_tid != -1)
-    {
-        ret = tkill( thread->unix_pid, thread->unix_tid, sig );
-        if (ret == -1 && errno == ENOSYS) ret = kill( thread->unix_pid, sig );
-    }
-    if (ret == -1 && errno == ESRCH) /* thread got killed */
+    if (thread->unix_pid != -1)
     {
-        thread->unix_pid = -1;
-        thread->unix_tid = -1;
-    }
+        if (thread->unix_tid != -1)
+        {
+            ret = tkill( thread->unix_pid, thread->unix_tid, sig );
+            if (ret == -1 && errno == ENOSYS) ret = kill( thread->unix_pid, sig );
+        }
+        else ret = kill( thread->unix_pid, sig );
 
+        if (ret == -1 && errno == ESRCH) /* thread got killed */
+        {
+            thread->unix_pid = -1;
+            thread->unix_tid = -1;
+        }
+    }
     if (debug_level && ret != -1)
         fprintf( stderr, "%04x: *sent signal* %s\n", thread->id, get_signal_name( sig ));
     return (ret != -1);
@@ -314,20 +324,14 @@ static int read_thread_long( struct thread *thread, void *addr, unsigned long *d
     return 0;
 }
 
-/* write a long to a thread address space */
-static long write_thread_long( struct thread *thread, void *addr, unsigned long data, unsigned long mask )
+static int read_thread_int( struct thread *thread, void *addr, unsigned int *data )
 {
-    unsigned long old_data;
-    int res;
+    unsigned long long_data;
+    int ret;
 
-    if (mask != ~0ul)
-    {
-        if (read_thread_long( thread, addr, &old_data ) == -1) return -1;
-        data = (data & mask) | (old_data & ~mask);
-    }
-    if ((res = ptrace( PTRACE_POKEDATA, get_ptrace_pid(thread), (caddr_t)addr, data )) == -1)
-        file_set_error();
-    return res;
+    ret = read_thread_long( thread, addr, &long_data );
+    *data = long_data;
+    return ret;
 }
 
 /* return a thread of the process suitable for ptracing */
@@ -343,82 +347,21 @@ static struct thread *get_ptrace_thread( struct process *process )
     return NULL;
 }
 
-#ifdef HAVE_PROCESS_VM_READV
-static int read_process_memory_vm( struct thread *thread, client_ptr_t ptr, data_size_t size, char *dest )
-{
-    static int not_supported;
-    struct iovec local, remote;
-    ssize_t len;
-
-    if (not_supported) return -1;
-    if (thread->unix_pid == -1 || !is_process_init_done( thread->process ))
-    {
-        set_error( STATUS_ACCESS_DENIED );
-        return 0;
-    }
-
-    local.iov_len = remote.iov_len = size;
-    local.iov_base = dest;
-    remote.iov_base = (void *)(unsigned long)ptr;
-    len = process_vm_readv( thread->unix_pid, &local, 1, &remote, 1, 0 );
-    if (len < 0 && (errno == ENOSYS || errno == EPERM))
-    {
-        not_supported = 1;
-        return -1;
-    }
-    if (len == size) return 1;
-    set_error( len >= 0 ? STATUS_PARTIAL_COPY : STATUS_ACCESS_DENIED );
-    return 0;
-}
-#else
-static int read_process_memory_vm( struct thread *thread, client_ptr_t ptr, data_size_t size, char *dest )
+/* read data from a process memory space */
+int read_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, char *dest )
 {
-    return -1;
-}
-#endif
+    struct thread *thread = get_ptrace_thread( process );
+    unsigned int first_offset, last_offset, len;
+    unsigned long data, *addr;
 
-#ifdef HAVE_PROCESS_VM_WRITEV
-static int write_process_memory_vm( struct thread *thread, client_ptr_t ptr, data_size_t size, const char *src,
-                                    data_size_t *written )
-{
-    static int not_supported;
-    struct iovec local, remote;
-    ssize_t len;
+    if (!thread) return 0;
 
-    if (not_supported) return -1;
-    if (thread->unix_pid == -1 || !is_process_init_done( thread->process ))
+    if ((unsigned long)ptr != ptr)
     {
         set_error( STATUS_ACCESS_DENIED );
         return 0;
     }
 
-    local.iov_len = remote.iov_len = size;
-    local.iov_base = (void *)src;
-    remote.iov_base = (void *)(unsigned long)ptr;
-    len = process_vm_writev( thread->unix_pid, &local, 1, &remote, 1, 0 );
-    if (len < 0 && (errno == ENOSYS || errno == EPERM))
-    {
-        not_supported = 1;
-        return -1;
-    }
-
-    if (written) *written = max( len, 0 );
-    if (len != size) set_error( STATUS_PARTIAL_COPY );
-    return len == size;
-}
-#else
-static int write_process_memory_vm( struct thread *thread, client_ptr_t ptr, data_size_t size, const char *src,
-                                    data_size_t *written )
-{
-    return -1;
-}
-#endif
-
-static int read_process_memory_ptrace( struct thread *thread, client_ptr_t ptr, data_size_t size, char *dest )
-{
-    unsigned int first_offset, last_offset, len;
-    unsigned long data, *addr;
-
     first_offset = ptr % sizeof(long);
     last_offset = (size + first_offset) % sizeof(long);
     if (!last_offset) last_offset = sizeof(long);
@@ -433,7 +376,7 @@ static int read_process_memory_ptrace( struct thread *thread, client_ptr_t ptr,
             char procmem[24];
             int fd;
 
-            snprintf( procmem, sizeof(procmem), "/proc/%u/mem", thread->process->unix_pid );
+            snprintf( procmem, sizeof(procmem), "/proc/%u/mem", process->unix_pid );
             if ((fd = open( procmem, O_RDONLY )) != -1)
             {
                 ssize_t ret = pread( fd, dest, size, ptr );
@@ -471,11 +414,13 @@ static int read_process_memory_ptrace( struct thread *thread, client_ptr_t ptr,
     return !len;
 }
 
-/* read data from a process memory space */
-int read_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, char *dest )
+/* write data to a process memory space */
+int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, const char *src,
+                          data_size_t *written )
 {
     struct thread *thread = get_ptrace_thread( process );
-    int ret;
+    struct iovec local, remote;
+    ssize_t len;
 
     if (!thread) return 0;
 
@@ -485,123 +430,51 @@ int read_process_memory( struct process *process, client_ptr_t ptr, data_size_t
         return 0;
     }
 
-    if ((ret = read_process_memory_vm( thread, ptr, size, dest )) != -1) return ret;
-    return read_process_memory_ptrace( thread, ptr, size, dest );
-}
-
-/* make sure we can write to the whole address range */
-/* len is the total size (in longs) */
-static int check_process_write_access( struct thread *thread, long *addr, data_size_t len )
-{
-    size_t page = get_page_size() / sizeof(long);
-
-    for (;;)
+    if (thread->unix_pid == -1 || !is_process_init_done(thread->process))
     {
-        if (write_thread_long( thread, addr, 0, 0 ) == -1) return 0;
-        if (len <= page) break;
-        addr += page;
-        len -= page;
+        set_error( STATUS_ACCESS_DENIED );
+        return 0;
     }
-    return (write_thread_long( thread, addr + len - 1, 0, 0 ) != -1);
-}
-
-static int write_process_memory_ptrace( struct thread *thread, client_ptr_t ptr, data_size_t size, const char *src,
-                                        data_size_t *written )
-{
-    int ret = 0;
-    long data = 0;
-    data_size_t len;
-    long *addr;
-    unsigned long first_mask, first_offset, last_mask, last_offset;
-
-    /* compute the mask for the first long */
-    first_mask = ~0;
-    first_offset = ptr % sizeof(long);
-    memset( &first_mask, 0, first_offset );
-
-    /* compute the mask for the last long */
-    last_offset = (size + first_offset) % sizeof(long);
-    if (!last_offset) last_offset = sizeof(long);
-    last_mask = 0;
-    memset( &last_mask, 0xff, last_offset );
-
-    addr = (long *)(unsigned long)(ptr - first_offset);
-    len = (size + first_offset + sizeof(long) - 1) / sizeof(long);
-
-    if (suspend_for_ptrace( thread ))
+    local.iov_len = remote.iov_len = size;
+    local.iov_base = (void *)src;
+    remote.iov_base = (void *)(unsigned long)ptr;
+    len = process_vm_writev( thread->unix_pid, &local, 1, &remote, 1, 0);
+    if (written) *written = len == -1 ? 0 : len;
+    if (len == -1 || len != size)
     {
-        if (!check_process_write_access( thread, addr, len ))
-        {
-            set_error( STATUS_ACCESS_DENIED );
-            goto done;
-        }
-
-        if (len > 3)
-        {
-            char procmem[24];
-            int fd;
-
-            snprintf( procmem, sizeof(procmem), "/proc/%u/mem", thread->process->unix_pid );
-            if ((fd = open( procmem, O_WRONLY )) != -1)
-            {
-                ssize_t r = pwrite( fd, src, size, ptr );
-                close( fd );
-                if (r == size)
-                {
-                    ret = 1;
-                    goto done;
-                }
-            }
-        }
-
-        /* first word is special */
-        if (len > 1)
-        {
-            memcpy( (char *)&data + first_offset, src, sizeof(long) - first_offset );
-            src += sizeof(long) - first_offset;
-            if (write_thread_long( thread, addr++, data, first_mask ) == -1) goto done;
-            first_offset = 0;
-            len--;
-        }
-        else last_mask &= first_mask;
-
-        while (len > 1)
-        {
-            memcpy( &data, src, sizeof(long) );
-            src += sizeof(long);
-            if (write_thread_long( thread, addr++, data, ~0ul ) == -1) goto done;
-            len--;
-        }
-
-        /* last word is special too */
-        memcpy( (char *)&data + first_offset, src, last_offset - first_offset );
-        if (write_thread_long( thread, addr, data, last_mask ) == -1) goto done;
-        ret = 1;
-
-    done:
-        resume_after_ptrace( thread );
+        set_error( STATUS_PARTIAL_COPY );
+        return 0;
     }
-    if (ret && written) *written = size;
-    return ret;
+    return 1;
 }
 
-/* write data to a process memory space */
-int write_process_memory( struct process *process, client_ptr_t ptr, data_size_t size, const char *src,
-                          data_size_t *written )
+/* retrieve an LDT selector entry */
+void get_selector_entry( struct thread *thread, int entry, unsigned int *base,
+                         unsigned int *limit, unsigned char *flags )
 {
-    struct thread *thread = get_ptrace_thread( process );
-    int ret;
-
-    if (!thread) return 0;
-
-    if ((unsigned long)ptr != ptr)
+    if (!thread->process->ldt_copy)
     {
         set_error( STATUS_ACCESS_DENIED );
-        return 0;
+        return;
+    }
+    if (entry >= 8192)
+    {
+        set_error( STATUS_ACCESS_VIOLATION );
+        return;
+    }
+    if (suspend_for_ptrace( thread ))
+    {
+        unsigned int flags_buf;
+        unsigned long addr = (unsigned long)thread->process->ldt_copy + (entry * 4);
+
+        if (read_thread_int( thread, (void *)addr, base ) == -1) goto done;
+        if (read_thread_int( thread, (void *)(addr + (8192 * 4)), limit ) == -1) goto done;
+        addr = (unsigned long)thread->process->ldt_copy + (2 * 8192 * 4) + (entry & ~3);
+        if (read_thread_int( thread, (void *)addr, &flags_buf ) == -1) goto done;
+        *flags = flags_buf >> (entry & 3) * 8;
+    done:
+        resume_after_ptrace( thread );
     }
-
-    if ((ret = write_process_memory_vm( thread, ptr, size, src, written )) != -1) return ret;
-    return write_process_memory_ptrace( thread, ptr, size, src, written );
 }
 
 
@@ -627,7 +500,7 @@ void init_thread_context( struct thread *thread )
 /* retrieve the thread x86 registers */
 void get_thread_context( struct thread *thread, struct context_data *context, unsigned int flags )
 {
-    int i;
+    int i, pid = get_ptrace_tid(thread);
     long data[8];
 
     /* all other regs are handled on the client side */
@@ -646,7 +519,7 @@ void get_thread_context( struct thread *thread, struct context_data *context, un
     {
         if (i == 4 || i == 5) continue;
         errno = 0;
-        data[i] = ptrace( PTRACE_PEEKUSER, thread->unix_tid, DR_OFFSET(i), 0 );
+        data[i] = ptrace( PTRACE_PEEKUSER, pid, DR_OFFSET(i), 0 );
         if ((data[i] == -1) && errno)
         {
             file_set_error();
@@ -683,7 +556,7 @@ done:
 /* set the thread x86 registers */
 void set_thread_context( struct thread *thread, const struct context_data *context, unsigned int flags )
 {
-    int pid = thread->unix_tid;
+    int pid = get_ptrace_tid( thread );
 
     /* all other regs are handled on the client side */
     assert( flags == SERVER_CTX_DEBUG_REGISTERS );
@@ -741,7 +614,7 @@ void init_thread_context( struct thread *thread )
         struct dbreg dbregs;
 
         memset( &dbregs, 0, sizeof(dbregs) );
-        ptrace( PTRACE_SETDBREGS, thread->unix_tid, (caddr_t)&dbregs, 0 );
+        ptrace( PTRACE_SETDBREGS, get_ptrace_tid( thread ), (caddr_t)&dbregs, 0 );
         resume_after_ptrace( thread );
     }
     thread->system_regs = 0;
@@ -750,6 +623,7 @@ void init_thread_context( struct thread *thread )
 /* retrieve the thread x86 registers */
 void get_thread_context( struct thread *thread, struct context_data *context, unsigned int flags )
 {
+    int pid = get_ptrace_tid(thread);
     struct dbreg dbregs;
 
     /* all other regs are handled on the client side */
@@ -757,7 +631,7 @@ void get_thread_context( struct thread *thread, struct context_data *context, un
 
     if (!suspend_for_ptrace( thread )) return;
 
-    if (ptrace( PTRACE_GETDBREGS, thread->unix_tid, (caddr_t) &dbregs, 0 ) == -1) file_set_error();
+    if (ptrace( PTRACE_GETDBREGS, pid, (caddr_t) &dbregs, 0 ) == -1) file_set_error();
     else
     {
 #ifdef DBREG_DRX
@@ -791,6 +665,7 @@ void get_thread_context( struct thread *thread, struct context_data *context, un
 /* set the thread x86 registers */
 void set_thread_context( struct thread *thread, const struct context_data *context, unsigned int flags )
 {
+    int pid = get_ptrace_tid(thread);
     struct dbreg dbregs;
 
     /* all other regs are handled on the client side */
@@ -827,7 +702,7 @@ void set_thread_context( struct thread *thread, const struct context_data *conte
     dbregs.dr6 = context->debug.i386_regs.dr6;
     dbregs.dr7 = context->debug.i386_regs.dr7;
 #endif
-    if (ptrace( PTRACE_SETDBREGS, thread->unix_tid, (caddr_t)&dbregs, 0 ) != -1)
+    if (ptrace( PTRACE_SETDBREGS, pid, (caddr_t)&dbregs, 0 ) != -1)
     {
         thread->system_regs |= SERVER_CTX_DEBUG_REGISTERS;
     }
