diff --git a/server/queue.c b/server/queue.c
index 3e8ab2ef48a995cfbd24e8eafaea54fe99ad7a05..8e37d6b0aadc163e3f5831ff1ce832fa303501c4 100644
--- a/server/queue.c
+++ b/server/queue.c
@@ -26,7 +26,6 @@
 #include <stdlib.h>
 #include <unistd.h>
 #include <poll.h>
-#include <limits.h>
 
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
@@ -45,10 +44,8 @@
 #include "process.h"
 #include "request.h"
 #include "user.h"
-
-#define QS_DRIVER       0x80000000
-#define QS_HARDWARE     0x40000000
-#define QS_INTERNAL     (QS_DRIVER | QS_HARDWARE)
+#include "esync.h"
+#include "fsync.h"
 
 #define WM_NCMOUSEFIRST WM_NCMOUSEMOVE
 #define WM_NCMOUSELAST  (WM_NCMOUSEFIRST+(WM_MOUSELAST-WM_MOUSEFIRST))
@@ -105,6 +102,16 @@ struct timer
     lparam_t        lparam;    /* lparam for message */
 };
 
+struct pointer_state
+{
+    unsigned int pointer_id;    /* id of pointer currently tracked, ~0u if none */
+    user_handle_t pointer_win;  /* pointer window */
+    int button_msg_sent;        /* mouse button message was already sent */
+    int x;                      /* last seen pointer x coordinate */
+    int y;                      /* last seen pointer y coordinate */
+    int cursor_pos_updated;     /* updated mouse cursor position since last client tracking request */
+};
+
 struct thread_input
 {
     struct object          obj;           /* object header */
@@ -112,16 +119,15 @@ struct thread_input
     int                    caret_hide;    /* caret hide count */
     int                    caret_state;   /* caret on/off state */
     struct list            msg_list;      /* list of hardware messages */
-    timeout_t              user_time;     /* time of last user input */
     unsigned char          desktop_keystate[256]; /* desktop keystate when keystate was synced */
-    input_shm_t           *shared;        /* thread input in session shared memory */
+    struct pointer_state   pointer_state;
+    const input_shm_t     *shared;        /* thread input in session shared memory */
 };
 
 struct msg_queue
 {
     struct object          obj;             /* object header */
     struct fd             *fd;              /* optional file descriptor to poll */
-    struct object         *sync;            /* sync object for wait/signal */
     int                    paint_count;     /* pending paint messages count */
     int                    hotkey_count;    /* pending hotkey messages count */
     int                    quit_message;    /* is there a pending quit message? */
@@ -137,8 +143,13 @@ struct msg_queue
     struct timeout_user   *timeout;         /* timeout for next timer to expire */
     struct thread_input   *input;           /* thread input descriptor */
     struct hook_table     *hooks;           /* hook table */
+    timeout_t              last_get_msg;    /* time of last get message call */
     int                    keystate_lock;   /* owns an input keystate lock */
-    queue_shm_t           *shared;          /* queue in session shared memory */
+    const queue_shm_t     *shared;          /* queue in session shared memory */
+    int                    esync_fd;        /* esync file descriptor (signalled on message) */
+    int                    esync_in_msgwait; /* our thread is currently waiting on us */
+    unsigned int           fsync_idx;
+    int                    fsync_in_msgwait; /* our thread is currently waiting on us */
 };
 
 struct hotkey
@@ -152,7 +163,12 @@ struct hotkey
 };
 
 static void msg_queue_dump( struct object *obj, int verbose );
-static struct object *msg_queue_get_sync( struct object *obj );
+static int msg_queue_add_queue( struct object *obj, struct wait_queue_entry *entry );
+static void msg_queue_remove_queue( struct object *obj, struct wait_queue_entry *entry );
+static int msg_queue_signaled( struct object *obj, struct wait_queue_entry *entry );
+static int msg_queue_get_esync_fd( struct object *obj, enum esync_type *type );
+static unsigned int msg_queue_get_fsync_idx( struct object *obj, enum fsync_type *type );
+static void msg_queue_satisfied( struct object *obj, struct wait_queue_entry *entry );
 static void msg_queue_destroy( struct object *obj );
 static void msg_queue_poll_event( struct fd *fd, int event );
 static void thread_input_dump( struct object *obj, int verbose );
@@ -164,13 +180,14 @@ static const struct object_ops msg_queue_ops =
     sizeof(struct msg_queue),  /* size */
     &no_type,                  /* type */
     msg_queue_dump,            /* dump */
-    NULL,                      /* add_queue */
-    NULL,                      /* remove_queue */
-    NULL,                      /* signaled */
-    NULL,                      /* satisfied */
+    msg_queue_add_queue,       /* add_queue */
+    msg_queue_remove_queue,    /* remove_queue */
+    msg_queue_signaled,        /* signaled */
+    msg_queue_get_esync_fd,    /* get_esync_fd */
+    msg_queue_get_fsync_idx,   /* get_fsync_idx */
+    msg_queue_satisfied,       /* satisfied */
     no_signal,                 /* signal */
     no_get_fd,                 /* get_fd */
-    msg_queue_get_sync,        /* get_sync */
     default_map_access,        /* map_access */
     default_get_sd,            /* get_sd */
     default_set_sd,            /* set_sd */
@@ -205,10 +222,11 @@ static const struct object_ops thread_input_ops =
     no_add_queue,                 /* add_queue */
     NULL,                         /* remove_queue */
     NULL,                         /* signaled */
+    NULL,                         /* get_esync_fd */
+    NULL,                         /* get_fsync_idx */
     NULL,                         /* satisfied */
     no_signal,                    /* signal */
     no_get_fd,                    /* get_fd */
-    default_get_sync,             /* get_sync */
     default_map_access,           /* map_access */
     default_get_sd,               /* get_sd */
     default_set_sd,               /* set_sd */
@@ -254,8 +272,8 @@ static struct thread_input *create_thread_input( struct thread *thread )
     if ((input = alloc_object( &thread_input_ops )))
     {
         list_init( &input->msg_list );
-        input->user_time = 0;
         input->shared = NULL;
+        input->pointer_state.pointer_id = ~0u;
 
         if (!(input->desktop = get_thread_desktop( thread, 0 /* FIXME: access rights */ )))
         {
@@ -265,7 +283,7 @@ static struct thread_input *create_thread_input( struct thread *thread )
         memcpy( input->desktop_keystate, (const void *)input->desktop->shared->keystate,
                 sizeof(input->desktop_keystate) );
 
-        if (!(input->shared = alloc_shared_object( sizeof(*input->shared) )))
+        if (!(input->shared = alloc_shared_object()))
         {
             release_object( input );
             return NULL;
@@ -308,7 +326,6 @@ static struct msg_queue *create_msg_queue( struct thread *thread, struct thread_
     if ((queue = alloc_object( &msg_queue_ops )))
     {
         queue->fd              = NULL;
-        queue->sync            = NULL;
         queue->paint_count     = 0;
         queue->hotkey_count    = 0;
         queue->quit_message    = 0;
@@ -318,15 +335,25 @@ static struct msg_queue *create_msg_queue( struct thread *thread, struct thread_
         queue->timeout         = NULL;
         queue->input           = (struct thread_input *)grab_object( input );
         queue->hooks           = NULL;
+        queue->last_get_msg    = current_time;
         queue->keystate_lock   = 0;
+        queue->esync_fd        = -1;
+        queue->esync_in_msgwait = 0;
+        queue->fsync_idx       = 0;
+        queue->fsync_in_msgwait = 0;
         list_init( &queue->send_result );
         list_init( &queue->callback_result );
         list_init( &queue->pending_timers );
         list_init( &queue->expired_timers );
         for (i = 0; i < NB_MSG_KINDS; i++) list_init( &queue->msg_list[i] );
 
-        if (!(queue->sync = create_internal_sync( 1, 0 ))) goto error;
-        if (!(queue->shared = alloc_shared_object( sizeof(*queue->shared) )))
+        if (do_fsync())
+            queue->fsync_idx = fsync_alloc_shm( 0, 0 );
+
+        if (do_esync())
+            queue->esync_fd = esync_create_fd( 0, 0 );
+
+        if (!(queue->shared = alloc_shared_object()))
         {
             release_object( queue );
             return NULL;
@@ -335,12 +362,10 @@ static struct msg_queue *create_msg_queue( struct thread *thread, struct thread_
         SHARED_WRITE_BEGIN( queue->shared, queue_shm_t )
         {
             memset( (void *)shared->hooks_count, 0, sizeof(shared->hooks_count) );
-            shared->access_time = monotonic_time;
             shared->wake_mask = 0;
             shared->wake_bits = 0;
             shared->changed_mask = 0;
             shared->changed_bits = 0;
-            shared->internal_bits = 0;
         }
         SHARED_WRITE_END;
 
@@ -354,10 +379,6 @@ static struct msg_queue *create_msg_queue( struct thread *thread, struct thread_
     }
     if (new_input) release_object( new_input );
     return queue;
-
-error:
-    release_object( queue );
-    return NULL;
 }
 
 /* free the message queue of a thread at thread exit */
@@ -372,8 +393,8 @@ void free_msg_queue( struct thread *thread )
 /* synchronize thread input keystate with the desktop */
 static void sync_input_keystate( struct thread_input *input )
 {
-    input_shm_t *input_shm = input->shared;
-    desktop_shm_t *desktop_shm;
+    const input_shm_t *input_shm = input->shared;
+    const desktop_shm_t *desktop_shm;
     int i;
 
     if (!input->desktop || input_shm->keystate_lock) return;
@@ -394,7 +415,7 @@ static void sync_input_keystate( struct thread_input *input )
 /* locks thread input keystate to prevent synchronization */
 static void lock_input_keystate( struct thread_input *input )
 {
-    input_shm_t *input_shm = input->shared;
+    const input_shm_t *input_shm = input->shared;
 
     SHARED_WRITE_BEGIN( input_shm, input_shm_t )
     {
@@ -406,7 +427,7 @@ static void lock_input_keystate( struct thread_input *input )
 /* unlock the thread input keystate and synchronize it again */
 static void unlock_input_keystate( struct thread_input *input )
 {
-    input_shm_t *input_shm = input->shared;
+    const input_shm_t *input_shm = input->shared;
 
     SHARED_WRITE_BEGIN( input_shm, input_shm_t )
     {
@@ -421,7 +442,7 @@ static void unlock_input_keystate( struct thread_input *input )
 static int assign_thread_input( struct thread *thread, struct thread_input *new_input )
 {
     struct msg_queue *queue = thread->queue;
-    input_shm_t *input_shm;
+    const input_shm_t *input_shm;
 
     if (!queue)
     {
@@ -485,7 +506,7 @@ static struct message *alloc_hardware_message( lparam_t info, struct hw_msg_sour
 
 static int is_cursor_clipped( struct desktop *desktop )
 {
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     struct rectangle top_rect, clip_rect = desktop_shm->cursor.clip;
     get_virtual_screen_rect( desktop, &top_rect, 1 );
     return !is_rect_equal( &clip_rect, &top_rect );
@@ -495,7 +516,7 @@ static void queue_cursor_message( struct desktop *desktop, user_handle_t win, un
                                   lparam_t wparam, lparam_t lparam )
 {
     static const struct hw_msg_source source = { IMDT_UNAVAILABLE, IMO_SYSTEM };
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     struct thread_input *input;
     struct message *msg;
 
@@ -532,7 +553,7 @@ static int update_desktop_cursor_window( struct desktop *desktop, user_handle_t
 
     if (updated && (input = get_desktop_cursor_thread_input( desktop )))
     {
-        input_shm_t *input_shm = input->shared;
+        const input_shm_t *input_shm = input->shared;
         user_handle_t handle = input_shm->cursor_count < 0 ? 0 : input_shm->cursor;
         /* when clipping send the message to the foreground window as well, as some driver have an artificial overlay window */
         if (is_cursor_clipped( desktop )) queue_cursor_message( desktop, 0, WM_WINE_SETCURSOR, win, handle );
@@ -544,7 +565,7 @@ static int update_desktop_cursor_window( struct desktop *desktop, user_handle_t
 
 static int update_desktop_cursor_pos( struct desktop *desktop, user_handle_t win, int x, int y )
 {
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     int updated;
     unsigned int time = get_tick_count();
 
@@ -602,7 +623,7 @@ static void set_cursor_pos( struct desktop *desktop, int x, int y )
 /* sync cursor position after window change */
 void update_cursor_pos( struct desktop *desktop )
 {
-    desktop_shm_t *desktop_shm;
+    const desktop_shm_t *desktop_shm;
 
     desktop_shm = desktop->shared;
     set_cursor_pos( desktop, desktop_shm->cursor.x, desktop_shm->cursor.y );
@@ -612,7 +633,7 @@ void update_cursor_pos( struct desktop *desktop )
 static void get_message_defaults( struct msg_queue *queue, int *x, int *y, unsigned int *time )
 {
     struct desktop *desktop = queue->input->desktop;
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
 
     *x = desktop_shm->cursor.x;
     *y = desktop_shm->cursor.y;
@@ -622,7 +643,7 @@ static void get_message_defaults( struct msg_queue *queue, int *x, int *y, unsig
 /* set the cursor clip rectangle */
 void set_clip_rectangle( struct desktop *desktop, const struct rectangle *rect, unsigned int flags, int reset )
 {
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     struct rectangle top_rect, new_rect;
     unsigned int old_flags;
     int x, y;
@@ -664,10 +685,9 @@ void set_clip_rectangle( struct desktop *desktop, const struct rectangle *rect,
 /* change the foreground input and reset the cursor clip rect */
 static void set_foreground_input( struct desktop *desktop, struct thread_input *input )
 {
-    input_shm_t *input_shm, *old_input_shm;
+    const input_shm_t *input_shm, *old_input_shm;
     shared_object_t dummy_obj = {0};
 
-    if (input) input->user_time = monotonic_time;
     if (desktop->foreground_input == input) return;
     input_shm = input ? input->shared : &dummy_obj.shm.input;
     old_input_shm = desktop->foreground_input ? desktop->foreground_input->shared : &dummy_obj.shm.input;
@@ -719,22 +739,18 @@ void add_queue_hook_count( struct thread *thread, unsigned int index, int count
 }
 
 /* check the queue status */
-static inline int get_queue_status( struct msg_queue *queue )
+static inline int is_signaled( struct msg_queue *queue )
 {
-    queue_shm_t *queue_shm = queue->shared;
+    const queue_shm_t *queue_shm = queue->shared;
     return (queue_shm->wake_bits & queue_shm->wake_mask) ||
-           (queue_shm->changed_bits & queue_shm->changed_mask) ||
-            queue_shm->internal_bits;
+           (queue_shm->changed_bits & queue_shm->changed_mask);
 }
 
 /* set some queue bits */
 static inline void set_queue_bits( struct msg_queue *queue, unsigned int bits )
 {
-    queue_shm_t *queue_shm = queue->shared;
-    unsigned int internal = bits & QS_INTERNAL;
-    bits &= ~QS_INTERNAL;
+    const queue_shm_t *queue_shm = queue->shared;
 
-    /* lock the key state on key press, including from hardware messages */
     if (bits & (QS_KEY | QS_MOUSEBUTTON))
     {
         if (!queue->keystate_lock) lock_input_keystate( queue->input );
@@ -745,36 +761,35 @@ static inline void set_queue_bits( struct msg_queue *queue, unsigned int bits )
     {
         shared->wake_bits |= bits;
         shared->changed_bits |= bits;
-        shared->internal_bits |= internal;
     }
     SHARED_WRITE_END;
 
-    if (get_queue_status( queue )) signal_sync( queue->sync );
+    if (is_signaled( queue )) wake_up( &queue->obj, 0 );
 }
 
 /* clear some queue bits */
 static inline void clear_queue_bits( struct msg_queue *queue, unsigned int bits )
 {
-    queue_shm_t *queue_shm = queue->shared;
-    unsigned int internal = bits & QS_INTERNAL;
-    bits &= ~QS_INTERNAL;
+    const queue_shm_t *queue_shm = queue->shared;
 
     SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
     {
         shared->wake_bits &= ~bits;
         shared->changed_bits &= ~bits;
-        shared->internal_bits &= ~internal;
-        bits = shared->wake_bits;
     }
     SHARED_WRITE_END;
 
-    /* release the keystate lock when last key message has been processed */
-    if (!internal && !(bits & (QS_KEY | QS_MOUSEBUTTON)))
+    if (do_fsync() && !is_signaled( queue ))
+        fsync_clear( &queue->obj );
+
+    if (do_esync() && !is_signaled( queue ))
+        esync_clear( queue->esync_fd );
+
+    if (!(queue_shm->wake_bits & (QS_KEY | QS_MOUSEBUTTON)))
     {
         if (queue->keystate_lock) unlock_input_keystate( queue->input );
         queue->keystate_lock = 0;
     }
-    if (!get_queue_status( queue )) reset_sync( queue->sync );
 }
 
 /* check if message is matched by the filter */
@@ -805,7 +820,8 @@ static inline int get_hardware_msg_bit( unsigned int message )
     if (message == WM_INPUT_DEVICE_CHANGE || message == WM_INPUT) return QS_RAWINPUT;
     if (message == WM_MOUSEMOVE || message == WM_NCMOUSEMOVE) return QS_MOUSEMOVE;
     if (message >= WM_KEYFIRST && message <= WM_KEYLAST) return QS_KEY;
-    if (message >= WM_WINE_FIRST_DRIVER_MSG && message <= WM_WINE_LAST_DRIVER_MSG) return QS_HARDWARE;
+    if (message == WM_WINE_CLIPCURSOR) return QS_RAWINPUT;
+    if (message == WM_WINE_SETCURSOR) return QS_RAWINPUT;
     return QS_MOUSEBUTTON;
 }
 
@@ -825,8 +841,8 @@ static inline unsigned int get_unique_id(void)
     return id;
 }
 
-/* lookup an already queued mouse message that matches the message, window and type */
-static struct message *find_mouse_message( struct thread_input *input, const struct message *msg )
+/* try to merge a WM_MOUSEMOVE message with the last in the list; return 1 if successful */
+static int merge_mousemove( struct thread_input *input, const struct message *msg )
 {
     struct message *prev;
     struct list *ptr;
@@ -837,28 +853,13 @@ static struct message *find_mouse_message( struct thread_input *input, const str
         if (prev->msg >> 31) continue; /* ignore internal messages */
         if (prev->msg != WM_INPUT) break;
     }
-    if (!ptr) return NULL;
-    if (prev->result) return NULL;
-    if (prev->win && msg->win && prev->win != msg->win) return NULL;
-    if (prev->msg != msg->msg) return NULL;
-    if (prev->type != msg->type) return NULL;
-    return prev;
-}
-
-/* try to merge a WM_MOUSEWHEEL message with the last in the list; return 1 if successful */
-static int merge_mousewheel( struct thread_input *input, const struct message *msg )
-{
-    struct message *prev;
-    int delta;
-
-    if (!(prev = find_mouse_message( input, msg ))) return 0;
-    if (prev->x != msg->x || prev->y != msg->y) return 0; /* don't merge if cursor has moved */
-
-    /* accumulate wheel delta */
-    delta = GET_WHEEL_DELTA_WPARAM(prev->wparam) + GET_WHEEL_DELTA_WPARAM(msg->wparam);
-    if (delta < SHRT_MIN || delta > SHRT_MAX) return 0;
-
-    prev->wparam  = MAKEWPARAM(GET_KEYSTATE_WPARAM(msg->wparam), delta);
+    if (!ptr) return 0;
+    if (prev->result) return 0;
+    if (prev->win && msg->win && prev->win != msg->win) return 0;
+    if (prev->msg != msg->msg) return 0;
+    if (prev->type != msg->type) return 0;
+    /* now we can merge it */
+    prev->wparam  = msg->wparam;
     prev->lparam  = msg->lparam;
     prev->x       = msg->x;
     prev->y       = msg->y;
@@ -869,50 +870,56 @@ static int merge_mousewheel( struct thread_input *input, const struct message *m
         struct hardware_msg_data *msg_data = msg->data;
         prev_data->info = msg_data->info;
     }
-    list_remove( &prev->entry );
-    list_add_tail( &input->msg_list, &prev->entry );
+    list_remove( ptr );
+    list_add_tail( &input->msg_list, ptr );
     return 1;
 }
 
-/* try to merge a WM_MOUSEMOVE message with the last in the list; return 1 if successful */
-static int merge_mousemove( struct thread_input *input, const struct message *msg )
+/* try to merge a unique message with the last in the list; return 1 if successful */
+static int merge_unique_message( struct thread_input *input, unsigned int message, const struct message *msg )
 {
     struct message *prev;
 
-    if (!(prev = find_mouse_message( input, msg ))) return 0;
+    LIST_FOR_EACH_ENTRY_REV( prev, &input->msg_list, struct message, entry )
+        if (prev->msg == message) break;
+    if (&prev->entry == &input->msg_list) return 0;
 
+    if (prev->result) return 0;
+    if (prev->win != msg->win) return 0;
+    if (prev->type != msg->type) return 0;
+
+    /* now we can merge it */
     prev->wparam  = msg->wparam;
     prev->lparam  = msg->lparam;
     prev->x       = msg->x;
     prev->y       = msg->y;
     prev->time    = msg->time;
-    if (msg->type == MSG_HARDWARE && prev->data && msg->data)
-    {
-        struct hardware_msg_data *prev_data = prev->data;
-        struct hardware_msg_data *msg_data = msg->data;
-        prev_data->info = msg_data->info;
-    }
     list_remove( &prev->entry );
     list_add_tail( &input->msg_list, &prev->entry );
+
     return 1;
 }
 
-/* try to merge a unique message with the last in the list; return 1 if successful */
-static int merge_unique_message( struct thread_input *input, unsigned int message, const struct message *msg )
+/* try to merge pointer update message with the last in the list; return 1 if successful */
+static int merge_pointerupdate( struct thread_input *input, const struct message *msg )
 {
     struct message *prev;
 
     LIST_FOR_EACH_ENTRY_REV( prev, &input->msg_list, struct message, entry )
-        if (prev->msg == message) break;
+    {
+        if (prev->msg == msg->msg) break;
+        if (prev->msg >= WM_POINTERUPDATE && prev->msg <= WM_POINTERLEAVE) return 0;
+    }
+
     if (&prev->entry == &input->msg_list) return 0;
 
     if (prev->result) return 0;
     if (prev->win != msg->win) return 0;
     if (prev->type != msg->type) return 0;
+    if (prev->wparam != msg->wparam) return 0;
+    if (prev->lparam != msg->lparam) return 0;
 
     /* now we can merge it */
-    prev->wparam  = msg->wparam;
-    prev->lparam  = msg->lparam;
     prev->x       = msg->x;
     prev->y       = msg->y;
     prev->time    = msg->time;
@@ -925,10 +932,10 @@ static int merge_unique_message( struct thread_input *input, unsigned int messag
 /* try to merge a message with the messages in the list; return 1 if successful */
 static int merge_message( struct thread_input *input, const struct message *msg )
 {
-    if (msg->msg == WM_MOUSEWHEEL) return merge_mousewheel( input, msg );
     if (msg->msg == WM_MOUSEMOVE) return merge_mousemove( input, msg );
     if (msg->msg == WM_WINE_CLIPCURSOR) return merge_unique_message( input, WM_WINE_CLIPCURSOR, msg );
     if (msg->msg == WM_WINE_SETCURSOR) return merge_unique_message( input, WM_WINE_SETCURSOR, msg );
+    if (msg->msg == WM_POINTERUPDATE) return merge_pointerupdate( input, msg );
     return 0;
 }
 
@@ -1283,31 +1290,110 @@ static void cleanup_results( struct msg_queue *queue )
 /* check if the thread owning the queue is hung (not checking for messages) */
 static int is_queue_hung( struct msg_queue *queue )
 {
-    /* queue is hung if it's signaled and thread didn't access it for more than 5 seconds */
-    return get_queue_status( queue ) && monotonic_time - queue->shared->access_time > 5 * TICKS_PER_SEC;
+    struct wait_queue_entry *entry;
+
+    if (current_time - queue->last_get_msg <= 5 * TICKS_PER_SEC)
+        return 0;  /* less than 5 seconds since last get message -> not hung */
+
+    LIST_FOR_EACH_ENTRY( entry, &queue->obj.wait_queue, struct wait_queue_entry, entry )
+    {
+        if (get_wait_queue_thread(entry)->queue == queue)
+            return 0;  /* thread is waiting on queue -> not hung */
+    }
+
+    if (do_fsync() && queue->fsync_in_msgwait)
+        return 0;   /* thread is waiting on queue in absentia -> not hung */
+
+    if (do_esync() && queue->esync_in_msgwait)
+        return 0;   /* thread is waiting on queue in absentia -> not hung */
+
+    return 1;
 }
 
-static struct object *msg_queue_get_sync( struct object *obj )
+static int msg_queue_add_queue( struct object *obj, struct wait_queue_entry *entry )
 {
     struct msg_queue *queue = (struct msg_queue *)obj;
-    assert( obj->ops == &msg_queue_ops );
-    return grab_object( queue->sync );
+
+    /* a thread can only wait on its own queue */
+    if (get_wait_queue_thread(entry)->queue != queue)
+    {
+        set_error( STATUS_ACCESS_DENIED );
+        return 0;
+    }
+
+    if (queue->fd && list_empty( &obj->wait_queue ))  /* first on the queue */
+        set_fd_events( queue->fd, POLLIN );
+    add_queue( obj, entry );
+    return 1;
+}
+
+static void msg_queue_remove_queue(struct object *obj, struct wait_queue_entry *entry )
+{
+    struct msg_queue *queue = (struct msg_queue *)obj;
+
+    remove_queue( obj, entry );
+    if (queue->fd && list_empty( &obj->wait_queue ))  /* last on the queue is gone */
+        set_fd_events( queue->fd, 0 );
 }
 
 static void msg_queue_dump( struct object *obj, int verbose )
 {
     struct msg_queue *queue = (struct msg_queue *)obj;
-    queue_shm_t *queue_shm = queue->shared;
+    const queue_shm_t *queue_shm = queue->shared;
     fprintf( stderr, "Msg queue bits=%x mask=%x\n",
              queue_shm->wake_bits, queue_shm->wake_mask );
 }
 
+static int msg_queue_signaled( struct object *obj, struct wait_queue_entry *entry )
+{
+    struct msg_queue *queue = (struct msg_queue *)obj;
+    int ret = 0;
+
+    if (queue->fd)
+    {
+        if ((ret = check_fd_events( queue->fd, POLLIN )))
+            /* stop waiting on select() if we are signaled */
+            set_fd_events( queue->fd, 0 );
+        else if (!list_empty( &obj->wait_queue ))
+            /* restart waiting on poll() if we are no longer signaled */
+            set_fd_events( queue->fd, POLLIN );
+    }
+    return ret || is_signaled( queue );
+}
+
+static int msg_queue_get_esync_fd( struct object *obj, enum esync_type *type )
+{
+    struct msg_queue *queue = (struct msg_queue *)obj;
+    *type = ESYNC_QUEUE;
+    return queue->esync_fd;
+}
+
+static unsigned int msg_queue_get_fsync_idx( struct object *obj, enum fsync_type *type )
+{
+    struct msg_queue *queue = (struct msg_queue *)obj;
+    *type = FSYNC_QUEUE;
+    return queue->fsync_idx;
+}
+
+static void msg_queue_satisfied( struct object *obj, struct wait_queue_entry *entry )
+{
+    struct msg_queue *queue = (struct msg_queue *)obj;
+    const queue_shm_t *queue_shm = queue->shared;
+
+    SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
+    {
+        shared->wake_mask = 0;
+        shared->changed_mask = 0;
+    }
+    SHARED_WRITE_END;
+}
+
 static void msg_queue_destroy( struct object *obj )
 {
     struct msg_queue *queue = (struct msg_queue *)obj;
     struct list *ptr;
     struct hotkey *hotkey, *hotkey2;
-    input_shm_t *input_shm = queue->input->shared;
+    const input_shm_t *input_shm = queue->input->shared;
     int i;
 
     cleanup_results( queue );
@@ -1345,7 +1431,8 @@ static void msg_queue_destroy( struct object *obj )
     if (queue->hooks) release_object( queue->hooks );
     if (queue->fd) release_object( queue->fd );
     if (queue->shared) free_shared_object( queue->shared );
-    if (queue->sync) release_object( queue->sync );
+    if (do_esync()) close( queue->esync_fd );
+    if (queue->fsync_idx) fsync_free_shm_idx( queue->fsync_idx );
 }
 
 static void msg_queue_poll_event( struct fd *fd, int event )
@@ -1355,13 +1442,13 @@ static void msg_queue_poll_event( struct fd *fd, int event )
 
     if (event & (POLLERR | POLLHUP)) set_fd_events( fd, -1 );
     else set_fd_events( queue->fd, 0 );
-    set_queue_bits( queue, QS_DRIVER );
+    wake_up( &queue->obj, 0 );
 }
 
 static void thread_input_dump( struct object *obj, int verbose )
 {
     struct thread_input *input = (struct thread_input *)obj;
-    input_shm_t *input_shm = input->shared;
+    const input_shm_t *input_shm = input->shared;
     fprintf( stderr, "Thread input focus=%08x capture=%08x active=%08x\n",
              input_shm->focus, input_shm->capture, input_shm->active );
 }
@@ -1384,7 +1471,7 @@ static void thread_input_destroy( struct object *obj )
 static inline void thread_input_cleanup_window( struct msg_queue *queue, user_handle_t window )
 {
     struct thread_input *input = queue->input;
-    input_shm_t *input_shm = input->shared;
+    const input_shm_t *input_shm = input->shared;
 
     SHARED_WRITE_BEGIN( input_shm, input_shm_t )
     {
@@ -1418,6 +1505,16 @@ static int check_queue_input_window( struct msg_queue *queue, user_handle_t wind
     return ret;
 }
 
+/* check if the thread queue is idle and set the process idle event if so */
+void check_thread_queue_idle( struct thread *thread )
+{
+    struct msg_queue *queue = thread->queue;
+    const queue_shm_t *queue_shm = queue->shared;
+
+    if ((queue_shm->wake_mask & QS_SMRESULT)) return;
+    if (thread->process->idle_event) set_event( thread->process->idle_event );
+}
+
 /* make sure the specified thread has a queue */
 int init_thread_queue( struct thread *thread )
 {
@@ -1446,7 +1543,7 @@ int attach_thread_input( struct thread *thread_from, struct thread *thread_to )
 
     if (thread_from->queue)
     {
-        input_shm_t *old_input_shm, *input_shm;
+        const input_shm_t *old_input_shm, *input_shm;
         old_input = thread_from->queue->input;
         old_input_shm = old_input->shared;
         input_shm = input->shared;
@@ -1462,7 +1559,7 @@ int attach_thread_input( struct thread *thread_from, struct thread *thread_to )
     ret = assign_thread_input( thread_from, input );
     if (ret)
     {
-        input_shm_t *input_shm = input->shared;
+        const input_shm_t *input_shm = input->shared;
         SHARED_WRITE_BEGIN( input_shm, input_shm_t )
         {
             memset( (void *)shared->keystate, 0, sizeof(shared->keystate) );
@@ -1482,7 +1579,7 @@ void detach_thread_input( struct thread *thread_from )
 
     if ((input = create_thread_input( thread_from )))
     {
-        input_shm_t *old_input_shm, *input_shm;
+        const input_shm_t *old_input_shm, *input_shm;
         old_input_shm = old_input->shared;
         input_shm = input->shared;
 
@@ -1554,14 +1651,20 @@ static void set_next_timer( struct msg_queue *queue )
 static struct timer *find_timer( struct msg_queue *queue, user_handle_t win,
                                  unsigned int msg, lparam_t id )
 {
-    struct timer *timer;
+    struct list *ptr;
 
     /* we need to search both lists */
 
-    LIST_FOR_EACH_ENTRY( timer, &queue->pending_timers, struct timer, entry )
+    LIST_FOR_EACH( ptr, &queue->pending_timers )
+    {
+        struct timer *timer = LIST_ENTRY( ptr, struct timer, entry );
         if (timer->win == win && timer->msg == msg && timer->id == id) return timer;
-    LIST_FOR_EACH_ENTRY( timer, &queue->expired_timers, struct timer, entry )
+    }
+    LIST_FOR_EACH( ptr, &queue->expired_timers )
+    {
+        struct timer *timer = LIST_ENTRY( ptr, struct timer, entry );
         if (timer->win == win && timer->msg == msg && timer->id == id) return timer;
+    }
     return NULL;
 }
 
@@ -1614,10 +1717,11 @@ static struct timer *find_expired_timer( struct msg_queue *queue, user_handle_t
                                          unsigned int get_first, unsigned int get_last,
                                          int remove )
 {
-    struct timer *timer;
+    struct list *ptr;
 
-    LIST_FOR_EACH_ENTRY( timer, &queue->expired_timers, struct timer, entry )
+    LIST_FOR_EACH( ptr, &queue->expired_timers )
     {
+        struct timer *timer = LIST_ENTRY( ptr, struct timer, entry );
         if (win && timer->win != win) continue;
         if (check_msg_filter( timer->msg, get_first, get_last ))
         {
@@ -1719,7 +1823,7 @@ static void update_key_state( volatile unsigned char *keystate, unsigned int msg
 
 static void update_thread_input_key_state( struct thread_input *input, unsigned int msg, lparam_t wparam )
 {
-    input_shm_t *input_shm = input->shared;
+    const input_shm_t *input_shm = input->shared;
     SHARED_WRITE_BEGIN( input_shm, input_shm_t )
     {
         update_key_state( shared->keystate, msg, wparam, 0 );
@@ -1769,7 +1873,7 @@ static void release_hardware_message( struct msg_queue *queue, unsigned int hw_i
 
 static int queue_hotkey_message( struct desktop *desktop, struct message *msg )
 {
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     struct hotkey *hotkey;
     unsigned int modifiers = 0;
 
@@ -1810,7 +1914,7 @@ static user_handle_t find_hardware_message_window( struct desktop *desktop, stru
                                                    struct message *msg, unsigned int *msg_code,
                                                    struct thread **thread )
 {
-    input_shm_t *input_shm = input ? input->shared : NULL;
+    const input_shm_t *input_shm = input ? input->shared : NULL;
     user_handle_t win = 0;
 
     *thread = NULL;
@@ -1819,7 +1923,6 @@ static user_handle_t find_hardware_message_window( struct desktop *desktop, stru
     {
     case QS_POINTER:
     case QS_RAWINPUT:
-    case QS_HARDWARE:
         if (!(win = msg->win) && input) win = input_shm->focus;
         break;
     case QS_KEY:
@@ -1884,19 +1987,19 @@ static unsigned int get_rawinput_device_flags( struct process *process, struct m
 /* queue a hardware message into a given thread input */
 static void queue_hardware_message( struct desktop *desktop, struct message *msg, int always_queue )
 {
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     user_handle_t win;
     struct thread *thread;
     struct thread_input *input;
     struct hardware_msg_data *msg_data = msg->data;
     unsigned int msg_code;
-    int flags, msg_bit;
+    int flags;
 
     update_desktop_key_state( desktop, msg->msg, msg->wparam );
     last_input_time = get_tick_count();
     if (msg->msg != WM_MOUSEMOVE) always_queue = 1;
 
-    switch ((msg_bit = get_hardware_msg_bit( msg->msg )))
+    switch (get_hardware_msg_bit( msg->msg ))
     {
     case QS_KEY:
         if (queue_hotkey_message( desktop, msg )) return;
@@ -1930,10 +2033,6 @@ static void queue_hardware_message( struct desktop *desktop, struct message *msg
 
     win = find_hardware_message_window( desktop, input, msg, &msg_code, &thread );
     flags = thread ? get_rawinput_device_flags( thread->process, msg ) : 0;
-    if (thread) input = thread->queue->input;
-    if (input && (get_hardware_msg_bit( msg->msg ) & (QS_KEY | QS_MOUSEBUTTON)))
-        input->user_time = monotonic_time;
-
     if (!win || !thread || (flags & RIDEV_NOLEGACY))
     {
         if (input && !(flags & RIDEV_NOLEGACY)) update_thread_input_key_state( input, msg->msg, msg->wparam );
@@ -1942,13 +2041,15 @@ static void queue_hardware_message( struct desktop *desktop, struct message *msg
         return;
     }
 
+    input = thread->queue->input;
+
     if (win != msg->win) always_queue = 1;
     if (!always_queue || merge_message( input, msg )) free_message( msg );
     else
     {
         msg->unique_id = 0;  /* will be set once we return it to the app */
         list_add_tail( &input->msg_list, &msg->entry );
-        set_queue_bits( thread->queue, msg_bit );
+        set_queue_bits( thread->queue, get_hardware_msg_bit( msg->msg ) );
     }
     release_object( thread );
 }
@@ -1998,25 +2099,13 @@ static struct thread *get_foreground_thread( struct desktop *desktop, user_handl
     /* if desktop has no foreground process, assume the receiving window is */
     if (desktop->foreground_input)
     {
-        input_shm_t *input_shm = desktop->foreground_input->shared;
-        if (!(window = input_shm->focus)) window = input_shm->active;
+        const input_shm_t *input_shm = desktop->foreground_input->shared;
+        return get_window_thread( input_shm->focus );
     }
     if (window) return get_window_thread( window );
     return NULL;
 }
 
-static int is_current_process_foreground( struct desktop *desktop )
-{
-    struct thread *thread;
-    int ret;
-
-    if (!(thread = get_foreground_thread( desktop, 0 ))) return 1;
-    ret = thread->process == current->process || thread->process->id == current->process->parent_id;
-    release_object( thread );
-
-    return ret;
-}
-
 /* user32 reserves 1 & 2 for winemouse and winekeyboard,
  * keep this in sync with user_private.h */
 #define WINE_MOUSE_HANDLE 1
@@ -2042,7 +2131,9 @@ static void rawmouse_init( struct rawinput *header, RAWMOUSE *rawmouse, int x, i
     header->wparam = 0;
     header->usage  = MAKELONG(HID_USAGE_GENERIC_MOUSE, HID_USAGE_PAGE_GENERIC);
 
-    rawmouse->usFlags       = MOUSE_MOVE_RELATIVE;
+    rawmouse->usFlags       = 0;
+    if (flags & MOUSEEVENTF_ABSOLUTE)    rawmouse->usFlags |= MOUSE_MOVE_ABSOLUTE;
+    if (flags & MOUSEEVENTF_VIRTUALDESK) rawmouse->usFlags |= MOUSE_VIRTUAL_DESKTOP;
     rawmouse->usButtonFlags = 0;
     rawmouse->usButtonData  = 0;
     for (i = 1; i < ARRAY_SIZE(button_flags); ++i)
@@ -2213,9 +2304,9 @@ static void dispatch_rawinput_message( struct desktop *desktop, struct rawinput_
 
 /* queue a hardware message for a mouse event */
 static int queue_mouse_message( struct desktop *desktop, user_handle_t win, const union hw_input *input,
-                                unsigned int origin, struct msg_queue *sender )
+                                unsigned int origin, struct msg_queue *sender, unsigned int send_flags )
 {
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     struct hardware_msg_data *msg_data;
     struct rawinput_message raw_msg;
     struct message *msg;
@@ -2273,7 +2364,7 @@ static int queue_mouse_message( struct desktop *desktop, user_handle_t win, cons
         y = desktop_shm->cursor.y;
     }
 
-    if ((foreground = get_foreground_thread( desktop, win )))
+    if (!(send_flags & SEND_HWMSG_NO_RAW) && (foreground = get_foreground_thread( desktop, win )))
     {
         memset( &raw_msg, 0, sizeof(raw_msg) );
         raw_msg.foreground = foreground;
@@ -2281,13 +2372,15 @@ static int queue_mouse_message( struct desktop *desktop, user_handle_t win, cons
         raw_msg.time       = time;
         raw_msg.message    = WM_INPUT;
         raw_msg.flags      = flags;
-        rawmouse_init( &raw_msg.rawinput, &raw_msg.data.mouse, x - desktop_shm->cursor.x, y - desktop_shm->cursor.y,
+        rawmouse_init( &raw_msg.rawinput, &raw_msg.data.mouse, input->mouse.x, input->mouse.y,
                        raw_msg.flags, input->mouse.data, input->mouse.info );
 
         dispatch_rawinput_message( desktop, &raw_msg );
         release_object( foreground );
     }
 
+    if (send_flags & SEND_HWMSG_NO_MSG) return 0;
+
     for (i = 0; i < ARRAY_SIZE( messages ); i++)
     {
         if (!messages[i]) continue;
@@ -2318,14 +2411,15 @@ static int queue_mouse_message( struct desktop *desktop, user_handle_t win, cons
 }
 
 static int queue_keyboard_message( struct desktop *desktop, user_handle_t win, const union hw_input *input,
-                                   unsigned int origin, struct msg_queue *sender, int repeat );
+                                   unsigned int origin, struct msg_queue *sender, unsigned int flags, int repeat );
 
 static void key_repeat_timeout( void *private )
 {
     struct desktop *desktop = private;
 
     desktop->key_repeat.timeout = NULL;
-    queue_keyboard_message( desktop, desktop->key_repeat.win, &desktop->key_repeat.input, IMO_HARDWARE, NULL, 1 );
+    queue_keyboard_message( desktop, desktop->key_repeat.win, &desktop->key_repeat.input,
+                            IMO_HARDWARE, NULL, desktop->key_repeat.flags, 1 );
 }
 
 static void stop_key_repeat( struct desktop *desktop )
@@ -2338,9 +2432,9 @@ static void stop_key_repeat( struct desktop *desktop )
 
 /* queue a hardware message for a keyboard event */
 static int queue_keyboard_message( struct desktop *desktop, user_handle_t win, const union hw_input *input,
-                                   unsigned int origin, struct msg_queue *sender, int repeat )
+                                   unsigned int origin, struct msg_queue *sender, unsigned int send_flags, int repeat )
 {
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     struct hw_msg_source source = { IMDT_KEYBOARD, origin };
     struct hardware_msg_data *msg_data;
     struct message *msg;
@@ -2454,6 +2548,7 @@ static int queue_keyboard_message( struct desktop *desktop, user_handle_t win, c
         {
             timeout_t timeout = repeat ? desktop->key_repeat.period : desktop->key_repeat.delay;
             desktop->key_repeat.input = *input;
+            desktop->key_repeat.flags = send_flags;
             desktop->key_repeat.input.kbd.time = 0;
             desktop->key_repeat.win = win;
             if (desktop->key_repeat.timeout) remove_timeout_user( desktop->key_repeat.timeout );
@@ -2461,7 +2556,7 @@ static int queue_keyboard_message( struct desktop *desktop, user_handle_t win, c
         }
     }
 
-    if (!unicode && (foreground = get_foreground_thread( desktop, win )))
+    if (!unicode && !(send_flags & SEND_HWMSG_NO_RAW) && (foreground = get_foreground_thread( desktop, win )))
     {
         struct rawinput_message raw_msg = {0};
         raw_msg.foreground = foreground;
@@ -2476,6 +2571,8 @@ static int queue_keyboard_message( struct desktop *desktop, user_handle_t win, c
         release_object( foreground );
     }
 
+    if (send_flags & SEND_HWMSG_NO_MSG) return 0;
+
     if (!(msg = alloc_hardware_message( input->kbd.info, source, time, 0 ))) return 0;
     msg_data = msg->data;
 
@@ -2511,8 +2608,126 @@ struct pointer
     user_handle_t win;
     int primary;
     union hw_input input;
+    int x;
+    int y;
 };
 
+static struct pointer *find_pointer_from_id( struct desktop *desktop, unsigned int id, int create )
+{
+    struct pointer *pointer;
+
+    LIST_FOR_EACH_ENTRY( pointer, &desktop->pointers, struct pointer, entry )
+        if (LOWORD(pointer->input.hw.wparam) == id) return pointer;
+
+    if (!create) return NULL;
+
+    pointer = mem_alloc( sizeof(struct pointer) );
+    pointer->timeout = NULL;
+    pointer->desktop = desktop;
+    pointer->primary = list_empty( &desktop->pointers );
+    pointer->x = 0;
+    pointer->y = 0;
+    list_add_tail( &desktop->pointers, &pointer->entry );
+
+    return pointer;
+}
+
+static void queue_mouse_message_from_pointer( struct desktop *desktop, unsigned int message, int x, int y,
+                                              user_handle_t pointer_win, unsigned int time )
+{
+    const struct hw_msg_source source = { IMDT_TOUCH, IMO_HARDWARE };
+    struct message *msg;
+
+    if (!(msg = alloc_hardware_message( 0xff515700, source, time, 0 ))) return;
+
+    msg->win       = get_user_full_handle( pointer_win );
+    msg->msg       = message;
+    msg->wparam    = 0;
+    msg->lparam    = 0;
+    msg->x         = x;
+    msg->y         = y;
+
+    if (!send_hook_ll_message( desktop, msg, WH_MOUSE_LL, 0, NULL ))
+        queue_hardware_message( desktop, msg, 1 );
+}
+
+static int track_mouse_message_from_pointer( struct thread_input *input, unsigned int pointer_id,
+                                              unsigned int input_message )
+{
+    struct pointer_state *state = &input->pointer_state;
+    struct desktop *desktop = input->desktop;
+    unsigned int time = get_tick_count();
+    const desktop_shm_t *desktop_shm = desktop->shared;
+    struct pointer *pointer;
+    int pointer_moved;
+
+    if (!input) return 0;
+    if (input_message == WM_POINTERUP)
+    {
+        /* hardware pointer might be destroyed at this moment so don't rely on its presence */
+        if (pointer_id != state->pointer_id) return 0;
+        if (!state->button_msg_sent)
+        {
+            queue_mouse_message_from_pointer( desktop, WM_MOUSEMOVE, state->x, state->y, state->pointer_win, time );
+            queue_mouse_message_from_pointer( desktop, WM_LBUTTONDOWN, state->x, state->y, state->pointer_win, time );
+            state->cursor_pos_updated = 1;
+        }
+        else if (state->x != desktop_shm->cursor.x || state->y != desktop_shm->cursor.y)
+        {
+            queue_mouse_message_from_pointer( desktop, WM_MOUSEMOVE, state->x, state->y, state->pointer_win, time );
+            state->cursor_pos_updated = 1;
+        }
+        queue_mouse_message_from_pointer( desktop, WM_LBUTTONUP, state->x, state->y, state->pointer_win, time );
+        state->pointer_id = ~0u;
+        return 1;
+    }
+
+    if (input_message != WM_POINTERDOWN && state->pointer_id != pointer_id) return 0;
+
+    if (!(pointer = find_pointer_from_id( desktop, pointer_id, 0 ))) return 0;
+    if (input_message == WM_POINTERDOWN)
+    {
+        if (state->pointer_id != pointer_id && list_count( &desktop->pointers ) > 1)
+        {
+            if (!state->button_msg_sent) state->pointer_id = ~0u;
+            return 0;
+        }
+        state->pointer_id = pointer_id;
+        state->pointer_win = pointer->win;
+        state->x = pointer->x;
+        state->y = pointer->y;
+        state->button_msg_sent = 0;
+        state->cursor_pos_updated = 0;
+        return 1;
+    }
+
+    if (input_message != WM_POINTERUPDATE) return 0;
+    state->pointer_win = pointer->win;
+
+    pointer_moved = state->x != pointer->x || state->y != pointer->y;
+
+    if (pointer_moved)
+    {
+        if (!state->button_msg_sent)
+        {
+            queue_mouse_message_from_pointer( desktop, WM_MOUSEMOVE, state->x, state->y, state->pointer_win, time );
+            /* TODO: emulate right button and double clicks. */
+            queue_mouse_message_from_pointer( desktop, WM_LBUTTONDOWN, state->x, state->y, state->pointer_win, time );
+            state->button_msg_sent = 1;
+            state->cursor_pos_updated = 1;
+        }
+        state->x = pointer->x;
+        state->y = pointer->y;
+    }
+    if (pointer_moved || (state->button_msg_sent
+                          && (state->x != desktop_shm->cursor.x || state->y != desktop_shm->cursor.y)))
+    {
+        queue_mouse_message_from_pointer( desktop, WM_MOUSEMOVE, state->x, state->y, state->pointer_win, time );
+        state->cursor_pos_updated = 1;
+    }
+    return 1;
+}
+
 static void queue_pointer_message( struct pointer *pointer, int repeated );
 
 static void pointer_message_timeout( void *private )
@@ -2531,51 +2746,36 @@ static void queue_pointer_message( struct pointer *pointer, int repeated )
     };
     struct hw_msg_source source = { IMDT_UNAVAILABLE, IMDT_TOUCH };
     struct desktop *desktop = pointer->desktop;
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     const union hw_input *input = &pointer->input;
     unsigned int i, wparam = input->hw.wparam;
-    timeout_t time = get_tick_count();
-    user_handle_t win = pointer->win;
     struct rectangle top_rect;
     struct message *msg;
-    int x, y;
 
     get_virtual_screen_rect( desktop, &top_rect, 0 );
-    x = LOWORD(input->hw.lparam) * (top_rect.right - top_rect.left) / 65535;
-    y = HIWORD(input->hw.lparam) * (top_rect.bottom - top_rect.top) / 65535;
+    pointer->x = LOWORD(input->hw.lparam) * (top_rect.right - top_rect.left) / 65535;
+    pointer->y = HIWORD(input->hw.lparam) * (top_rect.bottom - top_rect.top) / 65535;
 
     if (pointer->primary) wparam |= POINTER_MESSAGE_FLAG_PRIMARY << 16;
 
     for (i = 0; i < 2 && messages[input->hw.msg - WM_POINTERUPDATE][i]; i++)
     {
-        if (!(msg = alloc_hardware_message( 0, source, time, 0 ))) return;
+        if (!(msg = alloc_hardware_message( 0, source, get_tick_count(), 0 ))) return;
 
-        msg->win       = get_user_full_handle( win );
+        msg->win       = get_user_full_handle( pointer->win );
         msg->msg       = messages[input->hw.msg - WM_POINTERUPDATE][i];
         msg->wparam    = wparam;
-        msg->lparam    = MAKELONG(x, y);
+        msg->lparam    = MAKELONG(pointer->x, pointer->y);
         msg->x         = desktop_shm->cursor.x;
         msg->y         = desktop_shm->cursor.y;
 
         queue_hardware_message( desktop, msg, 1 );
     }
 
-    if (!repeated && pointer->primary && (msg = alloc_hardware_message( 0xff515700, source, time, 0 )))
-    {
-        unsigned int message = WM_MOUSEMOVE;
-        if (input->hw.msg == WM_POINTERDOWN) message = WM_LBUTTONDOWN;
-        else if (input->hw.msg == WM_POINTERUP) message = WM_LBUTTONUP;
-
-        msg->win       = get_user_full_handle( win );
-        msg->msg       = message;
-        msg->wparam    = 0;
-        msg->lparam    = 0;
-        msg->x         = x;
-        msg->y         = y;
-
-        if (!send_hook_ll_message( desktop, msg, WH_MOUSE_LL, 0, NULL ))
-            queue_hardware_message( desktop, msg, 0 );
-    }
+    if (desktop->foreground_input
+        && desktop->foreground_input->pointer_state.pointer_id == LOWORD(pointer->input.hw.wparam)
+        && desktop->foreground_input->pointer_state.button_msg_sent)
+        track_mouse_message_from_pointer( desktop->foreground_input, LOWORD(pointer->input.hw.wparam), input->hw.msg );
 
     if (input->hw.msg != WM_POINTERUP)
     {
@@ -2590,27 +2790,11 @@ static void queue_pointer_message( struct pointer *pointer, int repeated )
     }
 }
 
-static struct pointer *find_pointer_from_id( struct desktop *desktop, unsigned int id )
-{
-    struct pointer *pointer;
-
-    LIST_FOR_EACH_ENTRY( pointer, &desktop->pointers, struct pointer, entry )
-        if (LOWORD(pointer->input.hw.wparam) == id) return pointer;
-
-    pointer = mem_alloc( sizeof(struct pointer) );
-    pointer->timeout = NULL;
-    pointer->desktop = desktop;
-    pointer->primary = list_empty( &desktop->pointers );
-    list_add_tail( &desktop->pointers, &pointer->entry );
-
-    return pointer;
-}
-
 /* queue a hardware message for a custom type of event */
 static void queue_custom_hardware_message( struct desktop *desktop, user_handle_t win,
                                            unsigned int origin, const union hw_input *input )
 {
-    desktop_shm_t *desktop_shm = desktop->shared;
+    const desktop_shm_t *desktop_shm = desktop->shared;
     struct hw_msg_source source = { IMDT_UNAVAILABLE, origin };
     struct thread *foreground;
     struct pointer *pointer;
@@ -2640,7 +2824,7 @@ static void queue_custom_hardware_message( struct desktop *desktop, user_handle_
 
     if (input->hw.msg == WM_POINTERDOWN || input->hw.msg == WM_POINTERUP || input->hw.msg == WM_POINTERUPDATE)
     {
-        pointer = find_pointer_from_id( desktop, LOWORD(input->hw.wparam) );
+        pointer = find_pointer_from_id( desktop, LOWORD(input->hw.wparam), 1 );
         if (pointer->timeout) remove_timeout_user( pointer->timeout );
         pointer->input = *input;
         pointer->win = win;
@@ -2698,6 +2882,12 @@ static int check_hw_message_filter( user_handle_t win, unsigned int msg_code,
     }
 }
 
+/* is this message an internal driver notification message */
+static inline BOOL is_internal_hardware_message( unsigned int message )
+{
+    return (message >= WM_WINE_FIRST_DRIVER_MSG && message <= WM_WINE_LAST_DRIVER_MSG);
+}
+
 /* find a hardware message for the given queue */
 static int get_hardware_message( struct thread *thread, unsigned int hw_id, user_handle_t filter_win,
                                  unsigned int first, unsigned int last, unsigned int flags,
@@ -2728,14 +2918,13 @@ static int get_hardware_message( struct thread *thread, unsigned int hw_id, user
     }
 
     if (ptr == list_head( &input->msg_list ))
-        clear_bits = QS_INPUT | QS_HARDWARE;
+        clear_bits = QS_INPUT;
     else
         clear_bits = 0;  /* don't clear bits if we don't go through the whole list */
 
     while (ptr)
     {
         struct message *msg = LIST_ENTRY( ptr, struct message, entry );
-        int msg_bit = get_hardware_msg_bit( msg->msg );
         struct hardware_msg_data *data = msg->data;
 
         ptr = list_next( &input->msg_list, ptr );
@@ -2759,7 +2948,7 @@ static int get_hardware_message( struct thread *thread, unsigned int hw_id, user
             if (win_thread->queue->input == input)
             {
                 /* wake the other thread */
-                set_queue_bits( win_thread->queue, msg_bit );
+                set_queue_bits( win_thread->queue, get_hardware_msg_bit( msg->msg ) );
                 got_one = 1;
             }
             else
@@ -2778,7 +2967,7 @@ static int get_hardware_message( struct thread *thread, unsigned int hw_id, user
          * match the filter we skip it */
         if (got_one || !check_hw_message_filter( win, msg_code, filter_win, first, last ))
         {
-            clear_bits &= ~msg_bit;
+            clear_bits &= ~get_hardware_msg_bit( msg->msg );
             continue;
         }
 
@@ -2802,14 +2991,13 @@ static int get_hardware_message( struct thread *thread, unsigned int hw_id, user
 
         data->hw_id = msg->unique_id;
         set_reply_data( msg->data, msg->data_size );
-
-        if (msg_bit == QS_HARDWARE) flags |= PM_REMOVE; /* always remove internal hardware messages right away */
-        else if (!(msg_bit & (QS_RAWINPUT | QS_POINTER))) flags &= ~PM_REMOVE; /* wait for accept_hardware_message request */
-        if (flags & PM_REMOVE) release_hardware_message( current->queue, data->hw_id );
+        if ((get_hardware_msg_bit( msg->msg ) & (QS_RAWINPUT | QS_POINTER) && (flags & PM_REMOVE)) ||
+            is_internal_hardware_message( msg->msg ))
+            release_hardware_message( current->queue, data->hw_id );
         return 1;
     }
     /* nothing found, clear the hardware queue bits */
-    if (clear_bits) clear_queue_bits( thread->queue, clear_bits );
+    clear_queue_bits( thread->queue, clear_bits );
     return 0;
 }
 
@@ -3058,12 +3246,9 @@ DECL_HANDLER(is_window_hung)
 DECL_HANDLER(get_msg_queue_handle)
 {
     struct msg_queue *queue = get_current_queue();
-    struct process *process = current->process;
-    struct event *event;
 
     reply->handle = 0;
-    if (queue) reply->handle = alloc_handle( process, queue, SYNCHRONIZE, 0 );
-    if ((event = process->idle_event)) reply->idle_event = alloc_handle( process, event, GENERIC_ALL, 0 );
+    if (queue) reply->handle = alloc_handle( current->process, queue, SYNCHRONIZE, 0 );
 }
 
 
@@ -3092,10 +3277,7 @@ DECL_HANDLER(set_queue_fd)
     if ((unix_fd = get_file_unix_fd( file )) != -1)
     {
         if ((unix_fd = dup( unix_fd )) != -1)
-        {
             queue->fd = create_anonymous_fd( &msg_queue_fd_ops, unix_fd, &queue->obj, 0 );
-            set_fd_events( queue->fd, POLLIN );
-        }
         else
             file_set_error();
     }
@@ -3107,30 +3289,42 @@ DECL_HANDLER(set_queue_fd)
 DECL_HANDLER(set_queue_mask)
 {
     struct msg_queue *queue = get_current_queue();
-    queue_shm_t *queue_shm;
 
-    if (!queue) return;
-    queue_shm = queue->shared;
-
-    if (req->poll_events)
+    if (queue)
     {
-        if (!queue->fd) return;
-        clear_queue_bits( queue, QS_DRIVER );
-        set_fd_events( queue->fd, POLLIN );
-    }
+        const queue_shm_t *queue_shm = queue->shared;
 
-    SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
-    {
-        shared->access_time  = monotonic_time;
-        shared->wake_mask    = req->wake_mask;
-        shared->changed_mask = req->changed_mask;
-        reply->wake_bits     = shared->wake_bits;
-        reply->changed_bits  = shared->changed_bits;
-    }
-    SHARED_WRITE_END;
+        SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
+        {
+            shared->wake_mask = req->wake_mask;
+            shared->changed_mask = req->changed_mask;
+        }
+        SHARED_WRITE_END;
 
-    if (!get_queue_status( queue )) reset_sync( queue->sync );
-    else signal_sync( queue->sync );
+        reply->wake_bits    = queue_shm->wake_bits;
+        reply->changed_bits = queue_shm->changed_bits;
+
+        if (is_signaled( queue ))
+        {
+            /* if skip wait is set, do what would have been done in the subsequent wait */
+            if (req->skip_wait)
+            {
+                SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
+                {
+                    shared->wake_mask = 0;
+                    shared->changed_mask = 0;
+                }
+                SHARED_WRITE_END;
+            }
+            else wake_up( &queue->obj, 0 );
+        }
+
+        if (do_fsync() && !is_signaled( queue ))
+            fsync_clear( &queue->obj );
+
+        if (do_esync() && !is_signaled( queue ))
+            esync_clear( queue->esync_fd );
+    }
 }
 
 
@@ -3138,20 +3332,26 @@ DECL_HANDLER(set_queue_mask)
 DECL_HANDLER(get_queue_status)
 {
     struct msg_queue *queue = current->queue;
-    queue_shm_t *queue_shm;
+    if (queue)
+    {
+        const queue_shm_t *queue_shm = queue->shared;
 
-    if (!queue) return;
-    queue_shm = queue->shared;
+        if (do_fsync() && !is_signaled( queue ))
+            fsync_clear( &queue->obj );
 
-    SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
-    {
-        reply->wake_bits      = shared->wake_bits;
-        reply->changed_bits   = shared->changed_bits;
-        shared->changed_bits &= ~req->clear_bits;
-    }
-    SHARED_WRITE_END;
+        if (do_esync() && !is_signaled( queue ))
+            esync_clear( queue->esync_fd );
+
+        reply->wake_bits    = queue_shm->wake_bits;
+        reply->changed_bits = queue_shm->changed_bits;
 
-    if (!get_queue_status( queue )) reset_sync( queue->sync );
+        SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
+        {
+            shared->changed_bits &= ~req->clear_bits;
+        }
+        SHARED_WRITE_END;
+    }
+    else reply->wake_bits = reply->changed_bits = 0;
 }
 
 
@@ -3241,7 +3441,7 @@ DECL_HANDLER(send_hardware_message)
     struct desktop *desktop;
     unsigned int origin = (req->flags & SEND_HWMSG_INJECTED ? IMO_INJECTED : IMO_HARDWARE);
     struct msg_queue *sender = req->flags & SEND_HWMSG_INJECTED ? get_current_queue() : NULL;
-    desktop_shm_t *desktop_shm;
+    const desktop_shm_t *desktop_shm;
     int wait = 0;
 
     if (!(desktop = get_hardware_input_desktop( req->win ))) return;
@@ -3260,10 +3460,10 @@ DECL_HANDLER(send_hardware_message)
     switch (req->input.type)
     {
     case INPUT_MOUSE:
-        wait = queue_mouse_message( desktop, req->win, &req->input, origin, sender );
+        wait = queue_mouse_message( desktop, req->win, &req->input, origin, sender, req->flags );
         break;
     case INPUT_KEYBOARD:
-        wait = queue_keyboard_message( desktop, req->win, &req->input, origin, sender, 0 );
+        wait = queue_keyboard_message( desktop, req->win, &req->input, origin, sender, req->flags, 0 );
         break;
     case INPUT_HARDWARE:
         queue_custom_hardware_message( desktop, req->win, origin, &req->input );
@@ -3298,10 +3498,10 @@ DECL_HANDLER(get_message)
     struct list *ptr;
     struct msg_queue *queue = get_current_queue();
     user_handle_t get_win = get_user_full_handle( req->get_win );
-    queue_shm_t *queue_shm;
+    const queue_shm_t *queue_shm;
     unsigned int filter;
 
-    if (get_win && get_win != 1 && get_win != -1 && !get_user_object( get_win, NTUSER_OBJ_WINDOW ))
+    if (get_win && get_win != 1 && get_win != -1 && !get_user_object( get_win, USER_WINDOW ))
     {
         set_win32_error( ERROR_INVALID_WINDOW_HANDLE );
         return;
@@ -3321,11 +3521,7 @@ DECL_HANDLER(get_message)
         return;
     }
 
-    SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
-    {
-        shared->access_time = monotonic_time;
-    }
-    SHARED_WRITE_END;
+    queue->last_get_msg = current_time;
 
     /* first check for sent messages */
     if ((ptr = list_head( &queue->msg_list[SEND_MESSAGE] )))
@@ -3351,8 +3547,6 @@ DECL_HANDLER(get_message)
     }
     SHARED_WRITE_END;
 
-    if (!get_queue_status( queue )) reset_sync( queue->sync );
-
     /* then check for posted messages */
     if ((filter & QS_POSTMESSAGE) &&
         get_posted_message( queue, get_win, req->get_first, req->get_last, req->flags, reply ))
@@ -3398,9 +3592,13 @@ DECL_HANDLER(get_message)
         reply->wparam = timer->id;
         reply->lparam = timer->lparam;
         get_message_defaults( queue, &reply->x, &reply->y, &reply->time );
+        if (!(req->flags & PM_NOYIELD) && current->process->idle_event)
+            set_event( current->process->idle_event );
         return;
     }
 
+    if (get_win == -1 && current->process->idle_event) set_event( current->process->idle_event );
+
     SHARED_WRITE_BEGIN( queue_shm, queue_shm_t )
     {
         shared->wake_mask = req->wake_mask;
@@ -3408,9 +3606,14 @@ DECL_HANDLER(get_message)
     }
     SHARED_WRITE_END;
 
-    if (!get_queue_status( queue )) reset_sync( queue->sync );
-    else signal_sync( queue->sync );
     set_error( STATUS_PENDING );  /* FIXME */
+
+    if (do_fsync() && !is_signaled( queue ))
+        fsync_clear( &queue->obj );
+
+    if (do_esync() && !is_signaled( queue ))
+        esync_clear( queue->esync_fd );
+
 }
 
 
@@ -3773,7 +3976,7 @@ DECL_HANDLER(get_key_state)
     else
     {
         struct msg_queue *queue = get_current_queue();
-        input_shm_t *input_shm = queue->input->shared;
+        const input_shm_t *input_shm = queue->input->shared;
         unsigned char *keystate = (void *)input_shm->keystate;
         sync_input_keystate( queue->input );
         reply->state = keystate[req->key & 0xff];
@@ -3787,7 +3990,7 @@ DECL_HANDLER(set_key_state)
     struct msg_queue *queue = get_current_queue();
     struct desktop *desktop = queue->input->desktop;
     data_size_t size = min( 256, get_req_data_size() );
-    input_shm_t *input_shm = queue->input->shared;
+    const input_shm_t *input_shm = queue->input->shared;
 
     SHARED_WRITE_BEGIN( input_shm, input_shm_t )
     {
@@ -3816,35 +4019,26 @@ DECL_HANDLER(set_foreground_window)
 {
     struct thread *thread = NULL;
     struct desktop *desktop;
+    struct thread_input *input;
     struct msg_queue *queue = get_current_queue();
-    int is_desktop = 0, set_foreground = 0;
 
-    if (!queue || !(desktop = get_thread_desktop( current, 0 ))) return;
+    if (!(desktop = get_thread_desktop( current, 0 ))) return;
 
-    if (!(thread = make_window_foreground( desktop, req->handle, &is_desktop, &set_foreground )) ||
-        !thread->queue || !thread->queue->input)
-    {
-        set_win32_error( ERROR_INVALID_WINDOW_HANDLE );
-        goto done;
-    }
+    if (!(input = desktop->foreground_input)) reply->previous = 0;
+    else reply->previous = input->shared->active;
+
+    reply->send_msg_old = (reply->previous && desktop->foreground_input != queue->input);
+    reply->send_msg_new = FALSE;
 
-    if (set_foreground && !req->internal)
+    if (is_valid_foreground_window( req->handle ) &&
+        (thread = get_window_thread( req->handle )) &&
+        thread->queue->input->desktop == desktop)
     {
-        if (!current->process->set_foreground) current->process->set_foreground = 1;
-        else if (!is_current_process_foreground( desktop ) && queue->input && desktop->foreground_input &&
-                 queue->input->user_time < desktop->foreground_input->user_time)
-        {
-            set_error( STATUS_ACCESS_DENIED );
-            goto done;
-        }
+        set_foreground_input( desktop, thread->queue->input );
+        reply->send_msg_new = (desktop->foreground_input != queue->input);
     }
+    else set_win32_error( ERROR_INVALID_WINDOW_HANDLE );
 
-    reply->previous = desktop->foreground_input ? desktop->foreground_input->shared->active : 0;
-    reply->send_msg_old = (reply->previous && desktop->foreground_input != queue->input);
-    set_foreground_input( desktop, req->internal || !is_desktop ? thread->queue->input : NULL );
-    reply->send_msg_new = (desktop->foreground_input != queue->input);
-
-done:
     if (thread) release_object( thread );
     release_object( desktop );
 }
@@ -3858,7 +4052,7 @@ DECL_HANDLER(set_focus_window)
     reply->previous = 0;
     if (queue && check_queue_input_window( queue, req->handle ))
     {
-        input_shm_t *input_shm = queue->input->shared;
+        const input_shm_t *input_shm = queue->input->shared;
         SHARED_WRITE_BEGIN( input_shm, input_shm_t )
         {
             reply->previous = shared->focus;
@@ -3872,23 +4066,36 @@ DECL_HANDLER(set_focus_window)
 /* set the current thread active window */
 DECL_HANDLER(set_active_window)
 {
+    struct message *msg, *next;
     struct msg_queue *queue = get_current_queue();
+    struct desktop *desktop;
+
+    if (!(desktop = get_thread_desktop( current, 0 ))) return;
 
     reply->previous = 0;
     if (queue && check_queue_input_window( queue, req->handle ))
     {
         if (!req->handle || make_window_active( req->handle ))
         {
-            input_shm_t *input_shm = queue->input->shared;
+            const input_shm_t *input_shm = queue->input->shared;
+
             SHARED_WRITE_BEGIN( input_shm, input_shm_t )
             {
                 reply->previous = shared->active;
                 shared->active = get_user_full_handle( req->handle );
             }
             SHARED_WRITE_END;
+
+            if (desktop->foreground_input == queue->input && req->handle != reply->previous)
+            {
+                LIST_FOR_EACH_ENTRY_SAFE( msg, next, &queue->msg_list[POST_MESSAGE], struct message, entry )
+                    if (msg->msg == req->internal_msg) remove_queue_message( queue, msg, POST_MESSAGE );
+            }
         }
         else set_error( STATUS_INVALID_HANDLE );
     }
+
+    release_object( desktop );
 }
 
 
@@ -3901,7 +4108,7 @@ DECL_HANDLER(set_capture_window)
     if (queue && check_queue_input_window( queue, req->handle ))
     {
         struct thread_input *input = queue->input;
-        input_shm_t *input_shm = input->shared;
+        const input_shm_t *input_shm = input->shared;
 
         /* if in menu mode, reject all requests to change focus, except if the menu bit is set */
         if (input_shm->menu_owner && !(req->flags & CAPTURE_MENU))
@@ -3933,7 +4140,7 @@ DECL_HANDLER(set_caret_window)
     if (queue && check_queue_input_window( queue, req->handle ))
     {
         struct thread_input *input = queue->input;
-        input_shm_t *input_shm = input->shared;
+        const input_shm_t *input_shm = input->shared;
         user_handle_t caret = get_user_full_handle(req->handle);
 
         reply->previous  = input_shm->caret;
@@ -3956,7 +4163,7 @@ DECL_HANDLER(set_caret_window)
 DECL_HANDLER(set_caret_info)
 {
     struct msg_queue *queue = get_current_queue();
-    input_shm_t *input_shm;
+    const input_shm_t *input_shm;
     struct thread_input *input;
 
     if (!queue) return;
@@ -4015,9 +4222,9 @@ DECL_HANDLER(set_cursor)
     struct msg_queue *queue = get_current_queue();
     user_handle_t prev_cursor, new_cursor;
     struct thread_input *input;
-    input_shm_t *input_shm;
+    const input_shm_t *input_shm;
     struct desktop *desktop;
-    desktop_shm_t *desktop_shm;
+    const desktop_shm_t *desktop_shm;
 
     if (!queue) return;
     input = queue->input;
@@ -4032,7 +4239,7 @@ DECL_HANDLER(set_cursor)
     reply->prev_y      = desktop_shm->cursor.y;
 
     if ((req->flags & SET_CURSOR_HANDLE) && req->handle &&
-        !get_user_object( req->handle, NTUSER_OBJ_ICON ))
+        !get_user_object( req->handle, USER_CLIENT ))
     {
         set_win32_error( ERROR_INVALID_CURSOR_HANDLE );
         return;
@@ -4078,7 +4285,7 @@ DECL_HANDLER(get_rawinput_buffer)
 {
     const size_t align = is_machine_64bit( current->process->machine ) ? 7 : 3;
     data_size_t buffer_size = get_reply_max_size() & ~align;
-    struct msg_queue *queue = current->queue;
+    struct thread_input *input = current->queue->input;
     struct message *msg, *next_msg;
     int count = 0;
     char *buffer;
@@ -4091,7 +4298,7 @@ DECL_HANDLER(get_rawinput_buffer)
 
     if (!req->read_data)
     {
-        LIST_FOR_EACH_ENTRY( msg, &queue->input->msg_list, struct message, entry )
+        LIST_FOR_EACH_ENTRY( msg, &input->msg_list, struct message, entry )
         {
             if (msg->msg == WM_INPUT)
             {
@@ -4109,7 +4316,7 @@ DECL_HANDLER(get_rawinput_buffer)
 
         reply->next_size = get_reply_max_size();
 
-        LIST_FOR_EACH_ENTRY_SAFE( msg, next_msg, &queue->input->msg_list, struct message, entry )
+        LIST_FOR_EACH_ENTRY_SAFE( msg, next_msg, &input->msg_list, struct message, entry )
         {
             if (msg->msg == WM_INPUT)
             {
@@ -4139,7 +4346,6 @@ DECL_HANDLER(get_rawinput_buffer)
 
         if (!next_size)
         {
-            clear_queue_bits( queue, QS_RAWINPUT );
             if (count) next_size = sizeof(RAWINPUT);
             else reply->next_size = 0;
         }
@@ -4215,3 +4421,43 @@ DECL_HANDLER(set_keyboard_repeat)
 
     release_object( desktop );
 }
+
+DECL_HANDLER(track_mouse_from_pointer)
+{
+    struct msg_queue *queue = get_current_queue();
+
+    if (queue && check_queue_input_window( queue, req->win )
+        && track_mouse_message_from_pointer( queue->input, req->pointer_id, req->msg ))
+    {
+        reply->cursor_pos_updated = queue->input->pointer_state.cursor_pos_updated;
+        queue->input->pointer_state.cursor_pos_updated = 0;
+    }
+}
+
+DECL_HANDLER(esync_msgwait)
+{
+    struct msg_queue *queue = get_current_queue();
+
+    if (!queue) return;
+    queue->esync_in_msgwait = req->in_msgwait;
+
+    check_thread_queue_idle( current );
+
+    /* and start/stop waiting on the driver */
+    if (queue->fd)
+        set_fd_events( queue->fd, req->in_msgwait ? POLLIN : 0 );
+}
+
+DECL_HANDLER(fsync_msgwait)
+{
+    struct msg_queue *queue = get_current_queue();
+
+    if (!queue) return;
+    queue->fsync_in_msgwait = req->in_msgwait;
+
+    check_thread_queue_idle( current );
+
+    /* and start/stop waiting on the driver */
+    if (queue->fd)
+        set_fd_events( queue->fd, req->in_msgwait ? POLLIN : 0 );
+}
