diff --git a/server/signal.c b/server/signal.c
index 078951af6e418e347b0733344cc53c88ae937e85..802b7f936b9f51c8d065ca000d7cc4106ba791d4 100644
--- a/server/signal.c
+++ b/server/signal.c
@@ -62,10 +62,11 @@ static const struct object_ops handler_ops =
     no_add_queue,             /* add_queue */
     NULL,                     /* remove_queue */
     NULL,                     /* signaled */
+    NULL,                     /* get_esync_fd */
+    NULL,                     /* get_fsync_idx */
     NULL,                     /* satisfied */
     no_signal,                /* signal */
     no_get_fd,                /* get_fd */
-    default_get_sync,         /* get_sync */
     default_map_access,       /* map_access */
     default_get_sd,           /* get_sd */
     default_set_sd,           /* set_sd */
