diff --git a/server/sock.c b/server/sock.c
index 7785d3c7706fcae68ed438e11832c0e8a5e73264..73409c5a1dffe275e4210fded6ca4a9d42ed5553 100644
--- a/server/sock.c
+++ b/server/sock.c
@@ -87,14 +87,6 @@
 # define HAS_IRDA
 #endif
 
-#ifdef HAVE_BLUETOOTH_BLUETOOTH_H
-# include <bluetooth/bluetooth.h>
-# ifdef HAVE_BLUETOOTH_RFCOMM_H
-#  include <bluetooth/rfcomm.h>
-#  define HAS_BLUETOOTH
-# endif
-#endif
-
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
 #include "windef.h"
@@ -106,10 +98,6 @@
 #include "tcpmib.h"
 #include "wsipx.h"
 #include "af_irda.h"
-#include "bthsdpdef.h"
-#include "bluetoothapis.h"
-#include "bthdef.h"
-#include "ws2bth.h"
 #include "wine/afd.h"
 #include "wine/rbtree.h"
 
@@ -160,9 +148,6 @@ union unix_sockaddr
 #ifdef HAS_IRDA
     struct sockaddr_irda irda;
 #endif
-#ifdef HAS_BLUETOOTH
-    struct sockaddr_rc rfcomm;
-#endif
 };
 
 static struct list poll_list = LIST_INIT( poll_list );
@@ -483,13 +468,14 @@ static const struct object_ops sock_ops =
     sizeof(struct sock),          /* size */
     &file_type,                   /* type */
     sock_dump,                    /* dump */
-    NULL,                         /* add_queue */
-    NULL,                         /* remove_queue */
-    NULL,                         /* signaled */
-    NULL,                         /* satisfied */
+    add_queue,                    /* add_queue */
+    remove_queue,                 /* remove_queue */
+    default_fd_signaled,          /* signaled */
+    NULL,                         /* get_esync_fd */
+    NULL,                         /* get_fsync_idx */
+    no_satisfied,                 /* satisfied */
     no_signal,                    /* signal */
     sock_get_fd,                  /* get_fd */
-    default_fd_get_sync,          /* get_sync */
     default_map_access,           /* map_access */
     default_get_sd,               /* get_sd */
     default_set_sd,               /* set_sd */
@@ -585,21 +571,6 @@ static int sockaddr_from_unix( const union unix_sockaddr *uaddr, struct WS_socka
     }
 #endif
 
-#ifdef HAS_BLUETOOTH
-    case AF_BLUETOOTH:
-    {
-        SOCKADDR_BTH win = {0};
-        BLUETOOTH_ADDRESS addr = {0};
-
-        if (wsaddrlen < sizeof(win)) return -1;
-        win.addressFamily = WS_AF_BTH;
-
-        memcpy( addr.rgBytes, uaddr->rfcomm.rc_bdaddr.b, sizeof( addr.rgBytes ));
-        win.btAddr = addr.ullLong;
-        win.port = uaddr->rfcomm.rc_channel;
-        return sizeof(win);
-    }
-#endif
     case AF_UNSPEC:
         return 0;
 
@@ -679,24 +650,6 @@ static socklen_t sockaddr_to_unix( const struct WS_sockaddr *wsaddr, int wsaddrl
     }
 #endif
 
-#ifdef HAS_BLUETOOTH
-    case WS_AF_BTH:
-    {
-        SOCKADDR_BTH win = {0};
-        BLUETOOTH_ADDRESS addr = {0};
-
-        if (wsaddrlen != sizeof(win)) return 0;
-        memcpy( &win, wsaddr, sizeof(win) );
-        addr.ullLong = win.btAddr;
-
-        uaddr->rfcomm.rc_family = AF_BLUETOOTH;
-        memcpy( &uaddr->rfcomm.rc_bdaddr, addr.rgBytes, sizeof( addr.rgBytes ) );
-        /* There can only be a maximum of 30 RFCOMM channels, so UINT8_MAX is safe to use here. */
-        uaddr->rfcomm.rc_channel = win.port == BT_PORT_ANY ? UINT8_MAX : win.port;
-        return sizeof(uaddr->rfcomm);
-    }
-#endif
-
     case WS_AF_UNSPEC:
         switch (wsaddrlen)
         {
@@ -1843,9 +1796,6 @@ static int get_unix_family( int family )
 #endif
 #ifdef AF_IRDA
         case WS_AF_IRDA: return AF_IRDA;
-#endif
-#ifdef AF_BLUETOOTH
-        case WS_AF_BTH: return AF_BLUETOOTH;
 #endif
         case WS_AF_UNSPEC: return AF_UNSPEC;
         default: return -1;
@@ -1863,16 +1813,11 @@ static int get_unix_type( int type )
     }
 }
 
-static int get_unix_protocol( int family, int protocol )
+static int get_unix_protocol( int protocol )
 {
     if (protocol >= WS_NSPROTO_IPX && protocol <= WS_NSPROTO_IPX + 255)
         return protocol;
 
-#ifdef HAS_BLUETOOTH
-    if (family == WS_AF_BTH)
-        return protocol == WS_BTHPROTO_RFCOMM ? BTPROTO_RFCOMM : -1;
-#endif
-
     switch (protocol)
     {
         case WS_IPPROTO_ICMP: return IPPROTO_ICMP;
@@ -1926,13 +1871,11 @@ static int init_socket( struct sock *sock, int family, int type, int protocol )
 
     unix_family = get_unix_family( family );
     unix_type = get_unix_type( type );
-    unix_protocol = get_unix_protocol( family, protocol );
+    unix_protocol = get_unix_protocol( protocol );
 
     if (unix_protocol < 0)
     {
-        if (family && unix_family < 0)
-            set_win32_error( WSAEAFNOSUPPORT );
-        else if (type && unix_type < 0)
+        if (type && unix_type < 0)
             set_win32_error( WSAESOCKTNOSUPPORT );
         else
             set_win32_error( WSAEPROTONOSUPPORT );
@@ -1977,10 +1920,6 @@ static int init_socket( struct sock *sock, int family, int type, int protocol )
     if (sockfd == -1)
     {
         if (errno == EINVAL) set_win32_error( WSAESOCKTNOSUPPORT );
-#ifdef AF_BLUETOOTH
-        else if (errno == ESOCKTNOSUPPORT && unix_family == AF_BLUETOOTH)
-            set_win32_error( WSAEAFNOSUPPORT );
-#endif
         else set_win32_error( sock_get_error( errno ));
         return -1;
     }
@@ -3094,41 +3033,11 @@ static void sock_ioctl( struct fd *fd, ioctl_code_t code, struct async *async )
         if (check_addr_usage( sock, &bind_addr, v6only ))
             return;
 
-#ifdef HAS_BLUETOOTH
-        if (unix_addr.rfcomm.rc_family == AF_BLUETOOTH
-            && !(unix_addr.rfcomm.rc_channel >= 1 && unix_addr.rfcomm.rc_channel <= 30))
-        {
-            int i;
-            if (unix_addr.rfcomm.rc_channel != UINT8_MAX)
-            {
-                set_error( sock_get_ntstatus( EADDRNOTAVAIL ) );
-                return;
-            }
-            /* If the RFCOMM channel was set to BT_PORT_ANY, we need to find an available RFCOMM
-             *  channel. The Linux kernel has a similar mechanism, but the channel is only assigned
-             *  on listen(), which we cannot call yet. The other, albeit hacky/race-y way to find an available
-             *  channel is to loop through all valid channel values (1 to 30) until bind() succeeds.
-             */
-            for (i = 1; i <= 30; i++)
-            {
-                bind_addr.rfcomm.rc_channel = i;
-                if (!bind( unix_fd, &bind_addr.addr, unix_len ))
-                    break;
-                if (errno != EADDRINUSE)
-                {
-                    set_error( sock_get_ntstatus( errno ) );
-                    return;
-                }
-            }
-            if (i > 30)
-            {
-                set_error( sock_get_ntstatus( EADDRINUSE ) );
-                return;
-            }
-        }
-        else
-#endif
-        if (bind( unix_fd, &bind_addr.addr, unix_len ) < 0)
+        /* Quake (and similar family) fails if we can't bind to an IPX address. This often
+         * doesn't work on Linux, so just fake success. */
+        if (unix_addr.addr.sa_family == AF_IPX)
+            fprintf( stderr, "wine: HACK: Faking AF_IPX bind success.\n" );
+        else if (bind( unix_fd, &bind_addr.addr, unix_len ) < 0)
         {
             if (errno == EADDRINUSE && sock->reuseaddr)
                 errno = EACCES;
@@ -3706,10 +3615,11 @@ static const struct object_ops ifchange_ops =
     no_add_queue,            /* add_queue */
     NULL,                    /* remove_queue */
     NULL,                    /* signaled */
+    NULL,                    /* get_esync_fd */
+    NULL,                    /* get_fsync_idx */
     no_satisfied,            /* satisfied */
     no_signal,               /* signal */
     ifchange_get_fd,         /* get_fd */
-    default_get_sync,        /* get_sync */
     default_map_access,      /* map_access */
     default_get_sd,          /* get_sd */
     default_set_sd,          /* set_sd */
@@ -3928,10 +3838,11 @@ static const struct object_ops socket_device_ops =
     no_add_queue,               /* add_queue */
     NULL,                       /* remove_queue */
     NULL,                       /* signaled */
+    NULL,                       /* get_esync_fd */
+    NULL,                       /* get_fsync_idx */
     no_satisfied,               /* satisfied */
     no_signal,                  /* signal */
     no_get_fd,                  /* get_fd */
-    default_get_sync,           /* get_sync */
     default_map_access,         /* map_access */
     default_get_sd,             /* get_sd */
     default_set_sd,             /* set_sd */
