diff --git a/server/symlink.c b/server/symlink.c
index 238dcfad1c731449b0bd2397679bc7c488130698..27a90d91dff024dbf074f1e5ab81fc9a317d8751 100644
--- a/server/symlink.c
+++ b/server/symlink.c
@@ -71,10 +71,11 @@ static const struct object_ops symlink_ops =
     no_add_queue,                 /* add_queue */
     NULL,                         /* remove_queue */
     NULL,                         /* signaled */
+    NULL,                         /* get_esync_fd */
+    NULL,                         /* get_fsync_idx */
     NULL,                         /* satisfied */
     no_signal,                    /* signal */
     no_get_fd,                    /* get_fd */
-    default_get_sync,             /* get_sync */
     default_map_access,           /* map_access */
     default_get_sd,               /* get_sd */
     default_set_sd,               /* set_sd */
@@ -178,7 +179,7 @@ struct object *create_obj_symlink( struct object *root, const struct unicode_str
     data_size_t len;
     WCHAR *target_name;
 
-    if (!(target_name = target->ops->get_full_name( target, ~0u, &len )))
+    if (!(target_name = target->ops->get_full_name( target, &len )))
     {
         set_error( STATUS_INVALID_PARAMETER );
         return NULL;
