diff --git a/server/thread.c b/server/thread.c
index e18114ef9bbac45c8ded09971f74379096b032cd..cdb0549f873379cd2978e948284a41691a804db4 100644
--- a/server/thread.c
+++ b/server/thread.c
@@ -40,18 +40,11 @@
 #ifdef HAVE_SYS_RESOURCE_H
 #include <sys/resource.h>
 #endif
-#ifdef __APPLE__
-#include <mach/mach_init.h>
-#include <mach/mach_time.h>
-#include <mach/mach_port.h>
-#include <mach/thread_act.h>
-#endif
 
 #include "ntstatus.h"
 #define WIN32_NO_STATUS
 #include "windef.h"
 #include "winternl.h"
-#include "ddk/wdm.h"
 
 #include "file.h"
 #include "handle.h"
@@ -60,6 +53,8 @@
 #include "request.h"
 #include "user.h"
 #include "security.h"
+#include "esync.h"
+#include "fsync.h"
 
 
 /* thread queues */
@@ -85,18 +80,16 @@ struct thread_wait
 struct thread_apc
 {
     struct object       obj;      /* object header */
-    struct object      *sync;     /* sync object for wait/signal */
     struct list         entry;    /* queue linked list */
     struct thread      *caller;   /* thread that queued this apc */
     struct object      *owner;    /* object that queued this apc */
-    struct reserve     *reserve;  /* reserve object associated with apc object */
     int                 executed; /* has it been executed by the client? */
     union apc_call      call;     /* call arguments */
     union apc_result    result;   /* call results once executed */
 };
 
 static void dump_thread_apc( struct object *obj, int verbose );
-static struct object *thread_apc_get_sync( struct object *obj );
+static int thread_apc_signaled( struct object *obj, struct wait_queue_entry *entry );
 static void thread_apc_destroy( struct object *obj );
 static void clear_apc_queue( struct list *queue );
 
@@ -105,13 +98,14 @@ static const struct object_ops thread_apc_ops =
     sizeof(struct thread_apc),  /* size */
     &no_type,                   /* type */
     dump_thread_apc,            /* dump */
-    NULL,                       /* add_queue */
-    NULL,                       /* remove_queue */
-    NULL,                       /* signaled */
-    NULL,                       /* satisfied */
+    add_queue,                  /* add_queue */
+    remove_queue,               /* remove_queue */
+    thread_apc_signaled,        /* signaled */
+    NULL,                       /* get_esync_fd */
+    NULL,                       /* get_fsync_idx */
+    no_satisfied,               /* satisfied */
     no_signal,                  /* signal */
     no_get_fd,                  /* get_fd */
-    thread_apc_get_sync,        /* get_sync */
     default_map_access,         /* map_access */
     default_get_sd,             /* get_sd */
     default_set_sd,             /* set_sd */
@@ -130,10 +124,9 @@ static const struct object_ops thread_apc_ops =
 
 struct context
 {
-    struct object           obj;        /* object header */
-    struct object          *sync;       /* sync object for wait/signal */
-    unsigned int            status;     /* status of the context */
-    struct context_data     regs[2];    /* context data */
+    struct object   obj;        /* object header */
+    unsigned int    status;     /* status of the context */
+    struct context_data regs[2];/* context data */
 };
 #define CTX_NATIVE  0  /* context for native machine */
 #define CTX_WOW     1  /* context if thread is inside WoW */
@@ -142,21 +135,21 @@ struct context
 static const unsigned int system_flags = SERVER_CTX_DEBUG_REGISTERS;
 
 static void dump_context( struct object *obj, int verbose );
-static struct object *context_get_sync( struct object *obj );
-static void context_destroy( struct object *obj );
+static int context_signaled( struct object *obj, struct wait_queue_entry *entry );
 
 static const struct object_ops context_ops =
 {
     sizeof(struct context),     /* size */
     &no_type,                   /* type */
     dump_context,               /* dump */
-    NULL,                       /* add_queue */
-    NULL,                       /* remove_queue */
-    NULL,                       /* signaled */
-    NULL,                       /* satisfied */
+    add_queue,                  /* add_queue */
+    remove_queue,               /* remove_queue */
+    context_signaled,           /* signaled */
+    NULL,                       /* get_esync_fd */
+    NULL,                       /* get_fsync_idx */
+    no_satisfied,               /* satisfied */
     no_signal,                  /* signal */
     no_get_fd,                  /* get_fd */
-    context_get_sync,           /* get_sync */
     default_map_access,         /* map_access */
     default_get_sd,             /* get_sd */
     default_set_sd,             /* set_sd */
@@ -167,7 +160,7 @@ static const struct object_ops context_ops =
     no_open_file,               /* open_file */
     no_kernel_obj_list,         /* get_kernel_obj_list */
     no_close_handle,            /* close_handle */
-    context_destroy,            /* destroy */
+    no_destroy                  /* destroy */
 };
 
 
@@ -189,7 +182,9 @@ struct type_descr thread_type =
 };
 
 static void dump_thread( struct object *obj, int verbose );
-static struct object *thread_get_sync( struct object *obj );
+static int thread_signaled( struct object *obj, struct wait_queue_entry *entry );
+static int thread_get_esync_fd( struct object *obj, enum esync_type *type );
+static unsigned int thread_get_fsync_idx( struct object *obj, enum fsync_type *type );
 static unsigned int thread_map_access( struct object *obj, unsigned int access );
 static void thread_poll_event( struct fd *fd, int event );
 static struct list *thread_get_kernel_obj_list( struct object *obj );
@@ -200,13 +195,14 @@ static const struct object_ops thread_ops =
     sizeof(struct thread),      /* size */
     &thread_type,               /* type */
     dump_thread,                /* dump */
-    NULL,                       /* add_queue */
-    NULL,                       /* remove_queue */
-    NULL,                       /* signaled */
-    NULL,                       /* satisfied */
+    add_queue,                  /* add_queue */
+    remove_queue,               /* remove_queue */
+    thread_signaled,            /* signaled */
+    thread_get_esync_fd,        /* get_esync_fd */
+    thread_get_fsync_idx,       /* get_fsync_idx */
+    no_satisfied,               /* satisfied */
     no_signal,                  /* signal */
     no_get_fd,                  /* get_fd */
-    thread_get_sync,            /* get_sync */
     thread_map_access,          /* map_access */
     default_get_sd,             /* get_sd */
     default_set_sd,             /* set_sd */
@@ -255,139 +251,26 @@ void init_threading(void)
     if (nice_limit < 0 && debug_level) fprintf(stderr, "wine: Using setpriority to control niceness in the [%d,%d] range\n", nice_limit, -nice_limit );
 }
 
-static void apply_thread_priority( struct thread *thread )
+static void apply_thread_priority( struct thread *thread, int base_priority )
 {
     int min = -nice_limit, max = nice_limit, range = max - min, niceness;
-    int effective_priority = get_effective_thread_priority( thread );
 
     if (nice_limit >= 0) return;
 
     /* FIXME: handle realtime priorities using SCHED_RR if possible */
-    if (effective_priority >= LOW_REALTIME_PRIORITY) effective_priority = LOW_REALTIME_PRIORITY - 1;
-    /* map an NT application band [1,15] priority to [-nice_limit, nice_limit] */
-    niceness = (min + (effective_priority - 1) * range / 14);
+    if (base_priority > THREAD_BASE_PRIORITY_LOWRT) base_priority = THREAD_BASE_PRIORITY_LOWRT;
+    /* map an NT application band [1,15] base priority to [-nice_limit, nice_limit] */
+    niceness = (min + (base_priority - 1) * range / 14);
     setpriority( PRIO_PROCESS, thread->unix_tid, niceness );
 }
 
-#elif defined(__APPLE__)
-static unsigned int mach_ticks_per_second;
-
-void init_threading(void)
-{
-    struct mach_timebase_info tb_info;
-    if (mach_timebase_info( &tb_info ) == KERN_SUCCESS)
-    {
-        mach_ticks_per_second = (tb_info.denom * 1000000000U) / tb_info.numer;
-    }
-    else
-    {
-        const unsigned int best_guess = 24000000U;
-        mach_ticks_per_second = best_guess;
-    }
-}
-
-static int get_mach_importance( int effective_priority )
-{
-    int min = -31, max = 32, range = max - min;
-    return min + (effective_priority - 1) * range / 14;
-}
-
-static void apply_thread_priority( struct thread *thread )
-{
-    kern_return_t kr;
-    mach_msg_type_name_t type;
-    int throughput_qos, latency_qos;
-    struct thread_extended_policy thread_extended_policy;
-    struct thread_precedence_policy thread_precedence_policy;
-    mach_port_t thread_port, process_port = thread->process->trace_data;
-    int effective_priority = get_effective_thread_priority( thread );
-
-    if (!process_port) return;
-    kr = mach_port_extract_right( process_port, thread->unix_tid,
-                                  MACH_MSG_TYPE_COPY_SEND, &thread_port, &type );
-    if (kr != KERN_SUCCESS) return;
-    /* effective priority 15 is for time-critical threads, so not compute-bound */
-    thread_extended_policy.timeshare = effective_priority > 14 ? 0 : 1;
-    thread_precedence_policy.importance = get_mach_importance( effective_priority );
-    /* adapted from the QoS table at xnu/osfmk/kern/thread_policy.c */
-    switch (thread->base_priority)
-    {
-    case THREAD_PRIORITY_IDLE: /* THREAD_QOS_MAINTENANCE */
-    case THREAD_PRIORITY_LOWEST: /* THREAD_QOS_BACKGROUND */
-        throughput_qos = THROUGHPUT_QOS_TIER_5;
-        latency_qos = LATENCY_QOS_TIER_3;
-        break;
-    case THREAD_PRIORITY_BELOW_NORMAL: /* THREAD_QOS_UTILITY */
-        throughput_qos = THROUGHPUT_QOS_TIER_2;
-        latency_qos = LATENCY_QOS_TIER_3;
-        break;
-    case THREAD_PRIORITY_NORMAL: /* THREAD_QOS_LEGACY */
-    case THREAD_PRIORITY_ABOVE_NORMAL: /* THREAD_QOS_USER_INITIATED */
-        throughput_qos = THROUGHPUT_QOS_TIER_1;
-        latency_qos = LATENCY_QOS_TIER_1;
-        break;
-    case THREAD_PRIORITY_HIGHEST: /* THREAD_QOS_USER_INTERACTIVE */
-        throughput_qos = THROUGHPUT_QOS_TIER_0;
-        latency_qos = LATENCY_QOS_TIER_0;
-        break;
-    case THREAD_PRIORITY_TIME_CRITICAL:
-    default: /* THREAD_QOS_UNSPECIFIED */
-        throughput_qos = THROUGHPUT_QOS_TIER_UNSPECIFIED;
-        latency_qos = LATENCY_QOS_TIER_UNSPECIFIED;
-        break;
-    }
-    /* QOS_UNSPECIFIED is assigned the highest tier available, so it does not provide a limit */
-    if (effective_priority >= LOW_REALTIME_PRIORITY || effective_priority > thread->priority)
-    {
-        throughput_qos = THROUGHPUT_QOS_TIER_UNSPECIFIED;
-        latency_qos = LATENCY_QOS_TIER_UNSPECIFIED;
-    }
-
-    thread_policy_set( thread_port, THREAD_LATENCY_QOS_POLICY, (thread_policy_t)&latency_qos,
-                       THREAD_LATENCY_QOS_POLICY_COUNT );
-    thread_policy_set( thread_port, THREAD_THROUGHPUT_QOS_POLICY, (thread_policy_t)&throughput_qos,
-                       THREAD_THROUGHPUT_QOS_POLICY_COUNT );
-    thread_policy_set( thread_port, THREAD_EXTENDED_POLICY, (thread_policy_t)&thread_extended_policy,
-                       THREAD_EXTENDED_POLICY_COUNT );
-    thread_policy_set( thread_port, THREAD_PRECEDENCE_POLICY, (thread_policy_t)&thread_precedence_policy,
-                       THREAD_PRECEDENCE_POLICY_COUNT );
-    if (effective_priority >= LOW_REALTIME_PRIORITY)
-    {
-        /* For realtime threads we are requesting from the scheduler to be moved
-         * into the Mach realtime band (96-127) above the kernel.
-         * The scheduler will bump us back into the application band though if we
-         * lie too much about our computation constraints...
-         * The maximum available amount of resources granted in that band is using
-         * half of the available bus cycles, and computation (nominally 1/10 of
-         * the time constraint) is a hint to the scheduler where to place our
-         * realtime threads relative to each other.
-         * If someone is violating the time contraint policy, they will be moved
-         * back where they were (non-timeshare application band with very high
-         * importance), which is on XNU equivalent to setting SCHED_RR with the
-         * pthread API. */
-        struct thread_time_constraint_policy thread_time_constraint_policy;
-        int realtime_priority = effective_priority - LOW_REALTIME_PRIORITY + 1;
-        unsigned int max_constraint = mach_ticks_per_second / 2;
-        unsigned int max_computation = max_constraint / 10;
-        /* unfortunately we can't give a hint for the periodicity of calculations */
-        thread_time_constraint_policy.period = 0;
-        thread_time_constraint_policy.constraint = max_constraint;
-        thread_time_constraint_policy.computation = realtime_priority * max_computation / 16;
-        thread_time_constraint_policy.preemptible = effective_priority == HIGH_PRIORITY ? 0 : 1;
-        thread_policy_set( thread_port, THREAD_TIME_CONSTRAINT_POLICY,
-                           (thread_policy_t)&thread_time_constraint_policy,
-                           THREAD_TIME_CONSTRAINT_POLICY_COUNT );
-    }
-    mach_port_deallocate( mach_task_self(), thread_port );
-}
-
 #else
 
 void init_threading(void)
 {
 }
 
-static void apply_thread_priority( struct thread *thread )
+static void apply_thread_priority( struct thread *thread, int base_priority )
 {
 }
 
@@ -398,13 +281,14 @@ static inline void init_thread_structure( struct thread *thread )
 {
     int i;
 
-    thread->sync            = NULL;
-    thread->alert_sync      = NULL;
     thread->unix_pid        = -1;  /* not known yet */
     thread->unix_tid        = -1;  /* not known yet */
     thread->context         = NULL;
     thread->teb             = 0;
     thread->entry_point     = 0;
+    thread->esync_fd        = -1;
+    thread->esync_apc_fd    = -1;
+    thread->fsync_idx       = 0;
     thread->system_regs     = 0;
     thread->queue           = NULL;
     thread->wait            = NULL;
@@ -419,8 +303,6 @@ static inline void init_thread_structure( struct thread *thread )
     thread->state           = RUNNING;
     thread->exit_code       = 0;
     thread->priority        = 0;
-    thread->base_priority   = 0;
-    thread->disable_boost   = 0;
     thread->suspend         = 0;
     thread->dbg_hidden      = 0;
     thread->bypass_proc_suspend = 0;
@@ -434,7 +316,6 @@ static inline void init_thread_structure( struct thread *thread )
     thread->completion_wait = NULL;
 
     list_init( &thread->mutex_list );
-    list_init( &thread->d3dkmt_mutexes );
     list_init( &thread->system_apc );
     list_init( &thread->user_apc );
     list_init( &thread->kernel_object );
@@ -443,10 +324,9 @@ static inline void init_thread_structure( struct thread *thread )
         thread->inflight[i].server = thread->inflight[i].client = -1;
 }
 
-static inline int is_thread_suspended( struct thread *thread )
+static int get_effective_proc_suspend( struct thread *thread )
 {
-    if (thread->suspend) return 1;
-    return !thread->bypass_proc_suspend && thread->process->suspend;
+    return thread->bypass_proc_suspend ? 0 : thread->process->suspend;
 }
 
 /* check if address looks valid for a client-side data structure (TEB etc.) */
@@ -467,35 +347,20 @@ static void dump_context( struct object *obj, int verbose )
 }
 
 
-static struct object *context_get_sync( struct object *obj )
+static int context_signaled( struct object *obj, struct wait_queue_entry *entry )
 {
     struct context *context = (struct context *)obj;
-    assert( obj->ops == &context_ops );
-    return grab_object( context->sync );
+    return context->status != STATUS_PENDING;
 }
 
-static void context_destroy( struct object *obj )
-{
-    struct context *context = (struct context *)obj;
-    assert( obj->ops == &context_ops );
-    if (context->sync) release_object( context->sync );
-}
 
 static struct context *create_thread_context( struct thread *thread )
 {
     struct context *context;
     if (!(context = alloc_object( &context_ops ))) return NULL;
-    context->sync   = NULL;
     context->status = STATUS_PENDING;
     memset( &context->regs, 0, sizeof(context->regs) );
     context->regs[CTX_NATIVE].machine = native_machine;
-
-    if (!(context->sync = create_internal_sync( 1, 0 )))
-    {
-        release_object( context );
-        return NULL;
-    }
-
     return context;
 }
 
@@ -542,7 +407,6 @@ struct thread *create_thread( int fd, struct process *process, const struct secu
     thread->process = (struct process *)grab_object( process );
     thread->desktop = 0;
     thread->affinity = process->affinity;
-    thread->disable_boost = process->disable_boost;
     if (!current) current = thread;
 
     list_add_tail( &thread_list, &thread->entry );
@@ -562,9 +426,11 @@ struct thread *create_thread( int fd, struct process *process, const struct secu
         release_object( thread );
         return NULL;
     }
-    if (!(thread->request_fd = create_anonymous_fd( &thread_fd_ops, fd, &thread->obj, 0 ))) goto error;
-    if (!(thread->sync = create_internal_sync( 1, 0 ))) goto error;
-    if (get_inproc_device_fd() >= 0 && !(thread->alert_sync = create_inproc_internal_sync( 1, 0 ))) goto error;
+    if (!(thread->request_fd = create_anonymous_fd( &thread_fd_ops, fd, &thread->obj, 0 )))
+    {
+        release_object( thread );
+        return NULL;
+    }
 
     if (process->desktop)
     {
@@ -576,13 +442,23 @@ struct thread *create_thread( int fd, struct process *process, const struct secu
         }
     }
 
+    thread->fsync_idx = 0;
+
+    if (do_fsync())
+    {
+        thread->fsync_idx = fsync_alloc_shm( 0, 0 );
+        thread->fsync_apc_idx = fsync_alloc_shm( 0, 0 );
+    }
+
+    if (do_esync())
+    {
+        thread->esync_fd = esync_create_fd( 0, 0 );
+        thread->esync_apc_fd = esync_create_fd( 0, 0 );
+    }
+
     set_fd_events( thread->request_fd, POLLIN );  /* start listening to events */
     add_process_thread( thread->process, thread );
     return thread;
-
-error:
-    release_object( thread );
-    return NULL;
 }
 
 /* handle a client event */
@@ -614,7 +490,7 @@ static void cleanup_thread( struct thread *thread )
     if (thread->context)
     {
         thread->context->status = STATUS_ACCESS_DENIED;
-        signal_sync( thread->context->sync );
+        wake_up( &thread->context->obj, 0 );
         release_object( thread->context );
         thread->context = NULL;
     }
@@ -659,8 +535,14 @@ static void destroy_thread( struct object *obj )
     release_object( thread->process );
     if (thread->id) free_ptid( thread->id );
     if (thread->token) release_object( thread->token );
-    if (thread->alert_sync) release_object( thread->alert_sync );
-    if (thread->sync) release_object( thread->sync );
+
+    if (do_esync())
+        close( thread->esync_fd );
+    if (thread->fsync_idx)
+    {
+        fsync_free_shm_idx( thread->fsync_idx );
+        fsync_free_shm_idx( thread->fsync_apc_idx );
+    }
 }
 
 /* dump a thread on stdout for debugging purposes */
@@ -673,11 +555,24 @@ static void dump_thread( struct object *obj, int verbose )
              thread->id, thread->unix_pid, thread->unix_tid, thread->state );
 }
 
-static struct object *thread_get_sync( struct object *obj )
+static int thread_signaled( struct object *obj, struct wait_queue_entry *entry )
+{
+    struct thread *mythread = (struct thread *)obj;
+    return (mythread->state == TERMINATED);
+}
+
+static int thread_get_esync_fd( struct object *obj, enum esync_type *type )
 {
     struct thread *thread = (struct thread *)obj;
-    assert( obj->ops == &thread_ops );
-    return grab_object( thread->sync );
+    *type = ESYNC_MANUAL_SERVER;
+    return thread->esync_fd;
+}
+
+static unsigned int thread_get_fsync_idx( struct object *obj, enum fsync_type *type )
+{
+    struct thread *thread = (struct thread *)obj;
+    *type = FSYNC_MANUAL_SERVER;
+    return thread->fsync_idx;
 }
 
 static unsigned int thread_map_access( struct object *obj, unsigned int access )
@@ -696,11 +591,10 @@ static void dump_thread_apc( struct object *obj, int verbose )
     fprintf( stderr, "APC owner=%p type=%u\n", apc->owner, apc->call.type );
 }
 
-static struct object *thread_apc_get_sync( struct object *obj )
+static int thread_apc_signaled( struct object *obj, struct wait_queue_entry *entry )
 {
     struct thread_apc *apc = (struct thread_apc *)obj;
-    assert( obj->ops == &thread_apc_ops );
-    return grab_object( apc->sync );
+    return apc->executed;
 }
 
 static void thread_apc_destroy( struct object *obj )
@@ -716,8 +610,6 @@ static void thread_apc_destroy( struct object *obj )
             async_set_result( apc->owner, apc->call.async_io.status, 0 );
         release_object( apc->owner );
     }
-    if (apc->sync) release_object( apc->sync );
-    reserve_obj_unbind( apc->reserve );
 }
 
 /* queue an async procedure call */
@@ -727,22 +619,15 @@ static struct thread_apc *create_apc( struct object *owner, const union apc_call
 
     if ((apc = alloc_object( &thread_apc_ops )))
     {
-        apc->sync        = NULL;
         if (call_data) apc->call = *call_data;
         else apc->call.type = APC_NONE;
         apc->caller      = NULL;
         apc->owner       = owner;
-        apc->reserve     = NULL;
         apc->executed    = 0;
         apc->result.type = APC_NONE;
         if (owner) grab_object( owner );
-
-        if (!(apc->sync = create_internal_sync( 1, 0 )))
-        {
-            release_object( apc );
-            return NULL;
-        }
     }
+
     return apc;
 }
 
@@ -799,8 +684,19 @@ int set_thread_affinity( struct thread *thread, affinity_t affinity )
 
         CPU_ZERO( &set );
         for (i = 0, mask = 1; mask; i++, mask <<= 1)
-            if (affinity & mask) CPU_SET( i, &set );
-
+            if (affinity & mask)
+            {
+                if (thread->process->cpu_override.cpu_count)
+                {
+                    if (i >= thread->process->cpu_override.cpu_count)
+                        break;
+                    CPU_SET( thread->process->cpu_override.host_cpu_id[i], &set );
+                }
+                else
+                {
+                    CPU_SET( i, &set );
+                }
+            }
         ret = sched_setaffinity( thread->unix_tid, sizeof(set), &set );
     }
 #endif
@@ -818,82 +714,66 @@ affinity_t get_thread_affinity( struct thread *thread )
         unsigned int i;
 
         if (!sched_getaffinity( thread->unix_tid, sizeof(set), &set ))
+        {
             for (i = 0; i < 8 * sizeof(mask); i++)
-                if (CPU_ISSET( i, &set )) mask |= (affinity_t)1 << i;
+                if (CPU_ISSET( i, &set ))
+                {
+                    if (thread->process->cpu_override.cpu_count)
+                    {
+                        if (i < ARRAY_SIZE(thread->process->wine_cpu_id_from_host))
+                            mask |= (affinity_t)1 << thread->process->wine_cpu_id_from_host[i];
+                    }
+                    else
+                    {
+                        mask |= (affinity_t)1 << i;
+                    }
+                }
+        }
     }
 #endif
     if (!mask) mask = ~(affinity_t)0;
     return mask;
 }
 
-int get_effective_thread_priority( struct thread *thread )
-{
-    int priority = thread->priority;
-
-    if (thread->disable_boost || priority >= LOW_REALTIME_PRIORITY)
-        return priority;
-
-    if (get_process_first_thread( thread->process ) == thread)
-        priority++;
-
-    return min( priority, LOW_REALTIME_PRIORITY - 1 );
-}
-
-unsigned int set_thread_priority( struct thread *thread, int priority )
+static int get_base_priority( int priority_class, int priority )
 {
-    int priority_class = thread->process->priority;
-
-    if (priority < LOW_PRIORITY + 1 || priority > HIGH_PRIORITY) return STATUS_INVALID_PARAMETER;
-    if (priority_class != PROCESS_PRIOCLASS_REALTIME && priority >= LOW_REALTIME_PRIORITY) return STATUS_PRIVILEGE_NOT_HELD;
-
-    thread->priority = priority;
-
-    /* if thread is gone or hasn't started yet, this will be called again from init_thread with a unix_tid */
-    if (thread->state == RUNNING && thread->unix_tid != -1) apply_thread_priority( thread );
-
-    return STATUS_SUCCESS;
+    /* offsets taken from https://learn.microsoft.com/en-us/windows/win32/procthread/scheduling-priorities */
+    static const int class_offsets[] = { 4, 8, 13, 24, 6, 10 };
+    if (priority == THREAD_PRIORITY_IDLE) return (priority_class == PROCESS_PRIOCLASS_REALTIME ? 16 : 1);
+    if (priority == THREAD_PRIORITY_TIME_CRITICAL) return (priority_class == PROCESS_PRIOCLASS_REALTIME ? 31 : 15);
+    if (priority_class >= ARRAY_SIZE(class_offsets)) return 8;
+    return class_offsets[priority_class - 1] + priority;
 }
 
 #define THREAD_PRIORITY_REALTIME_HIGHEST 6
 #define THREAD_PRIORITY_REALTIME_LOWEST -7
 
-/* sets the thread base priority level, relative to its process base priority */
-unsigned int set_thread_base_priority( struct thread *thread, int base_priority )
+int set_thread_priority( struct thread *thread, int priority_class, int priority )
 {
-    int priority;
-    int is_realtime = thread->process->priority == PROCESS_PRIOCLASS_REALTIME;
+    int min = THREAD_PRIORITY_LOWEST, max = THREAD_PRIORITY_HIGHEST, base_priority;
 
-    if (base_priority != THREAD_PRIORITY_IDLE && base_priority != THREAD_PRIORITY_TIME_CRITICAL)
+    if (priority_class == PROCESS_PRIOCLASS_REALTIME)
     {
-        int min = is_realtime ? THREAD_PRIORITY_REALTIME_LOWEST : THREAD_PRIORITY_LOWEST;
-        int max = is_realtime ? THREAD_PRIORITY_REALTIME_HIGHEST : THREAD_PRIORITY_HIGHEST;
-
-        if (base_priority < min || base_priority > max)
-            return STATUS_INVALID_PARAMETER;
+        max = THREAD_PRIORITY_REALTIME_HIGHEST;
+        min = THREAD_PRIORITY_REALTIME_LOWEST;
     }
+    if ((priority < min || priority > max) &&
+        priority != THREAD_PRIORITY_IDLE &&
+        priority != THREAD_PRIORITY_TIME_CRITICAL)
+        return STATUS_INVALID_PARAMETER;
 
-    thread->base_priority = base_priority;
+    if (thread->state == TERMINATED)
+        return STATUS_THREAD_IS_TERMINATING;
 
-    switch (base_priority)
-    {
-    case THREAD_PRIORITY_IDLE:
-        priority = is_realtime ? LOW_REALTIME_PRIORITY : LOW_PRIORITY + 1;
-        break;
-    case THREAD_PRIORITY_TIME_CRITICAL:
-        priority = is_realtime ? HIGH_PRIORITY : LOW_REALTIME_PRIORITY - 1;
-        break;
-    default:
-        priority = thread->process->base_priority + base_priority;
-        break;
-    }
+    thread->priority = priority;
 
-    return set_thread_priority( thread, priority );
-}
+    /* if unix_tid == -1, thread is gone or hasn't started yet, this will be called again from init_thread with a unix_tid */
+    if (thread->unix_tid == -1)
+        return STATUS_SUCCESS;
 
-void set_thread_disable_boost( struct thread *thread, int disable_boost )
-{
-    thread->disable_boost = disable_boost;
-    apply_thread_priority( thread );
+    base_priority = get_base_priority( priority_class, priority );
+    apply_thread_priority( thread, base_priority );
+    return STATUS_SUCCESS;
 }
 
 /* set all information about a thread */
@@ -902,12 +782,7 @@ static void set_thread_info( struct thread *thread,
 {
     if (req->mask & SET_THREAD_INFO_PRIORITY)
     {
-        unsigned int status = set_thread_priority( thread, req->priority );
-        if (status) set_error( status );
-    }
-    if (req->mask & SET_THREAD_INFO_BASE_PRIORITY)
-    {
-        unsigned int status = set_thread_base_priority( thread, req->base_priority );
+        int status = set_thread_priority( thread, thread->process->priority, req->priority );
         if (status) set_error( status );
     }
     if (req->mask & SET_THREAD_INFO_AFFINITY)
@@ -925,8 +800,6 @@ static void set_thread_info( struct thread *thread,
         thread->entry_point = req->entry_point;
     if (req->mask & SET_THREAD_INFO_DBG_HIDDEN)
         thread->dbg_hidden = 1;
-    if (req->mask & SET_THREAD_INFO_DISABLE_BOOST)
-        set_thread_disable_boost( thread, req->disable_boost );
     if (req->mask & SET_THREAD_INFO_DESCRIPTION)
     {
         WCHAR *desc;
@@ -966,8 +839,11 @@ int suspend_thread( struct thread *thread )
     int old_count = thread->suspend;
     if (thread->suspend < MAXIMUM_SUSPEND_COUNT)
     {
-        if (!is_thread_suspended( thread )) stop_thread( thread );
-        thread->suspend++;
+        if (!(get_effective_proc_suspend( thread ) + thread->suspend++))
+        {
+            stop_thread( thread );
+            if (thread == current) return old_count | 0x80000000;
+        }
     }
     else set_error( STATUS_SUSPEND_COUNT_EXCEEDED );
     return old_count;
@@ -980,7 +856,7 @@ int resume_thread( struct thread *thread )
     if (thread->suspend > 0)
     {
         if (!(--thread->suspend)) resume_delayed_debug_events( thread );
-        if (!is_thread_suspended( thread )) wake_thread( thread );
+        if (!(thread->suspend + get_effective_proc_suspend( thread ))) wake_thread( thread );
     }
     return old_count;
 }
@@ -989,6 +865,7 @@ int resume_thread( struct thread *thread )
 int add_queue( struct object *obj, struct wait_queue_entry *entry )
 {
     grab_object( obj );
+    entry->obj = obj;
     list_add_tail( &obj->wait_queue, &entry->entry );
     return 1;
 }
@@ -1025,46 +902,6 @@ void set_wait_status( struct wait_queue_entry *entry, int status )
     entry->wait->status = status;
 }
 
-static void object_sync_satisfied( struct object *obj, struct wait_queue_entry *entry )
-{
-    struct object *sync = get_obj_sync( obj );
-    sync->ops->satisfied( sync, entry );
-    release_object( sync );
-}
-
-static void object_sync_remove_queue( struct object *obj, struct wait_queue_entry *entry )
-{
-    struct object *sync = get_obj_sync( obj );
-    sync->ops->remove_queue( sync, entry );
-    release_object( sync );
-}
-
-static int object_sync_add_queue( struct object *obj, struct wait_queue_entry *entry )
-{
-    struct object *sync = get_obj_sync( obj );
-    int ret = sync->ops->add_queue( sync, entry );
-    release_object( sync );
-    return ret;
-}
-
-static int object_sync_signaled( struct object *obj, struct wait_queue_entry *entry )
-{
-    struct object *sync = get_obj_sync( obj );
-    int ret = sync->ops->signaled( sync, entry );
-    release_object( sync );
-    return ret;
-}
-
-void signal_sync( struct object *obj )
-{
-    obj->ops->signal( obj, 0, 1 );
-}
-
-void reset_sync( struct object *obj )
-{
-    obj->ops->signal( obj, 0, 0 );
-}
-
 /* finish waiting */
 static unsigned int end_wait( struct thread *thread, unsigned int status )
 {
@@ -1081,22 +918,18 @@ static unsigned int end_wait( struct thread *thread, unsigned int status )
         if (wait->select == SELECT_WAIT_ALL)
         {
             for (i = 0, entry = wait->queues; i < wait->count; i++, entry++)
-                object_sync_satisfied( entry->obj, entry );
+                entry->obj->ops->satisfied( entry->obj, entry );
         }
         else
         {
             entry = wait->queues + status;
-            object_sync_satisfied( entry->obj, entry );
+            entry->obj->ops->satisfied( entry->obj, entry );
         }
         status = wait->status;
         if (wait->abandoned) status += STATUS_ABANDONED_WAIT_0;
     }
     for (i = 0, entry = wait->queues; i < wait->count; i++, entry++)
-    {
-        object_sync_remove_queue( entry->obj, entry );
-        release_object( entry->obj );
-        entry->obj = NULL;
-    }
+        entry->obj->ops->remove_queue( entry->obj, entry );
     if (wait->user) remove_timeout_user( wait->user );
     free( wait );
     return status;
@@ -1108,7 +941,7 @@ static int wait_on( const union select_op *select_op, unsigned int count, struct
 {
     struct thread_wait *wait;
     struct wait_queue_entry *entry;
-    unsigned int i;
+    unsigned int i, idle = 0;
 
     if (!(wait = mem_alloc( FIELD_OFFSET(struct thread_wait, queues[count]) ))) return 0;
     wait->next    = current->wait;
@@ -1126,16 +959,17 @@ static int wait_on( const union select_op *select_op, unsigned int count, struct
     {
         struct object *obj = objects[i];
         entry->wait = wait;
-        if (!object_sync_add_queue( obj, entry ))
+        if (!obj->ops->add_queue( obj, entry ))
         {
             wait->count = i;
             end_wait( current, get_error() );
             return 0;
         }
 
-        entry->obj = grab_object( obj );
+        if (obj == (struct object *)current->queue) idle = 1;
     }
 
+    if (idle) check_thread_queue_idle( current );
     return current->wait ? 1 : 0;
 }
 
@@ -1171,7 +1005,7 @@ static int check_wait( struct thread *thread )
         return STATUS_KERNEL_APC;
 
     /* Suspended threads may not acquire locks, but they can run system APCs */
-    if (is_thread_suspended( thread )) return -1;
+    if (get_effective_proc_suspend( thread ) + thread->suspend > 0) return -1;
 
     if (wait->select == SELECT_WAIT_ALL)
     {
@@ -1179,13 +1013,13 @@ static int check_wait( struct thread *thread )
         /* Note: we must check them all anyway, as some objects may
          * want to do something when signaled, even if others are not */
         for (i = 0, entry = wait->queues; i < wait->count; i++, entry++)
-            not_ok |= !object_sync_signaled( entry->obj, entry );
+            not_ok |= !entry->obj->ops->signaled( entry->obj, entry );
         if (!not_ok) return STATUS_WAIT_0;
     }
     else
     {
         for (i = 0, entry = wait->queues; i < wait->count; i++, entry++)
-            if (object_sync_signaled( entry->obj, entry )) return i;
+            if (entry->obj->ops->signaled( entry->obj, entry )) return i;
     }
 
     if ((wait->flags & SELECT_ALERTABLE) && !list_empty(&thread->user_apc)) return STATUS_USER_APC;
@@ -1258,7 +1092,7 @@ int wake_thread_queue_entry( struct wait_queue_entry *entry )
     client_ptr_t cookie;
 
     if (thread->wait != wait) return 0;  /* not the current wait */
-    if (is_thread_suspended( thread )) return 0;  /* cannot acquire locks */
+    if (get_effective_proc_suspend( thread ) + thread->suspend > 0) return 0;  /* cannot acquire locks */
 
     assert( wait->select != SELECT_WAIT_ALL );
 
@@ -1281,7 +1115,7 @@ static void thread_timeout( void *ptr )
 
     wait->user = NULL;
     if (thread->wait != wait) return; /* not the top-level wait, ignore it */
-    if (is_thread_suspended( thread )) return;  /* suspended, ignore it */
+    if (thread->suspend + get_effective_proc_suspend( thread ) > 0) return;  /* suspended, ignore it */
 
     if (debug_level) fprintf( stderr, "%04x: *wakeup* signaled=TIMEOUT\n", thread->id );
     end_wait( thread, STATUS_TIMEOUT );
@@ -1301,7 +1135,7 @@ static int signal_object( obj_handle_t handle )
     obj = get_handle_obj( current->process, handle, 0, NULL );
     if (obj)
     {
-        ret = obj->ops->signal( obj, get_handle_access( current->process, handle ), -1 );
+        ret = obj->ops->signal( obj, get_handle_access( current->process, handle ));
         release_object( obj );
     }
     return ret;
@@ -1392,6 +1226,12 @@ void wake_up( struct object *obj, int max )
     struct list *ptr;
     int ret;
 
+    if (do_fsync())
+        fsync_wake_up( obj );
+
+    if (do_esync())
+        esync_wake_up( obj );
+
     LIST_FOR_EACH( ptr, &obj->wait_queue )
     {
         struct wait_queue_entry *entry = LIST_ENTRY( ptr, struct wait_queue_entry, entry );
@@ -1419,7 +1259,8 @@ static inline struct list *get_apc_queue( struct thread *thread, enum apc_type t
 /* check if thread is currently waiting for a (system) apc */
 static inline int is_in_apc_wait( struct thread *thread )
 {
-    return (is_thread_suspended( thread ) || (thread->wait && (thread->wait->flags & SELECT_INTERRUPTIBLE)));
+    return (get_effective_proc_suspend( thread ) || thread->suspend ||
+            (thread->wait && (thread->wait->flags & SELECT_INTERRUPTIBLE)));
 }
 
 /* queue an existing APC to a given thread */
@@ -1476,9 +1317,13 @@ static int queue_apc( struct process *process, struct thread *thread, struct thr
     list_add_tail( queue, &apc->entry );
     if (!list_prev( queue, &apc->entry ))  /* first one */
     {
-        if (apc->call.type == APC_USER && thread->alert_sync)
-            signal_inproc_sync( thread->alert_sync );
         wake_thread( thread );
+
+        if (do_fsync() && queue == &thread->user_apc)
+            fsync_wake_futex( thread->fsync_apc_idx );
+
+        if (do_esync() && queue == &thread->user_apc)
+            esync_wake_fd( thread->esync_apc_fd );
     }
 
     return 1;
@@ -1509,10 +1354,8 @@ void thread_cancel_apc( struct thread *thread, struct object *owner, enum apc_ty
         if (apc->owner != owner) continue;
         list_remove( &apc->entry );
         apc->executed = 1;
-        signal_sync( apc->sync );
+        wake_up( &apc->obj, 0 );
         release_object( apc );
-        if (list_empty( &thread->user_apc ) && thread->alert_sync)
-            reset_inproc_sync( thread->alert_sync );
         return;
     }
 }
@@ -1527,9 +1370,14 @@ static struct thread_apc *thread_dequeue_apc( struct thread *thread, int system
     {
         apc = LIST_ENTRY( ptr, struct thread_apc, entry );
         list_remove( ptr );
-        if (list_empty( &thread->user_apc ) && thread->alert_sync)
-            reset_inproc_sync( thread->alert_sync );
     }
+
+    if (do_fsync() && list_empty( &thread->system_apc ) && list_empty( &thread->user_apc ))
+        fsync_clear_futex( thread->fsync_apc_idx );
+
+    if (do_esync() && list_empty( &thread->system_apc ) && list_empty( &thread->user_apc ))
+        esync_clear( thread->esync_apc_fd );
+
     return apc;
 }
 
@@ -1543,7 +1391,7 @@ static void clear_apc_queue( struct list *queue )
         struct thread_apc *apc = LIST_ENTRY( ptr, struct thread_apc, entry );
         list_remove( &apc->entry );
         apc->executed = 1;
-        signal_sync( apc->sync );
+        wake_up( &apc->obj, 0 );
         release_object( apc );
     }
 }
@@ -1625,7 +1473,11 @@ void kill_thread( struct thread *thread, int violent_death )
     }
     kill_console_processes( thread, 0 );
     abandon_mutexes( thread );
-    signal_sync( thread->sync );
+    if (do_fsync())
+        fsync_abandon_mutexes( thread );
+    if (do_esync())
+        esync_abandon_mutexes( thread );
+    wake_up( &thread->obj, 0 );
     if (violent_death) send_thread_signal( thread, SIGQUIT );
     cleanup_thread( thread );
     remove_process_thread( thread->process, thread );
@@ -1752,7 +1604,29 @@ static int init_thread( struct thread *thread, int reply_fd, int wait_fd )
 DECL_HANDLER(init_first_thread)
 {
     struct process *process = current->process;
-    int fd;
+    const struct cpu_topology_override *cpu_override = get_req_data();
+    unsigned int have_cpu_override = get_req_data_size() / sizeof(*cpu_override);
+    unsigned int i;
+
+    if (have_cpu_override)
+    {
+        if (cpu_override->cpu_count > ARRAY_SIZE(process->wine_cpu_id_from_host))
+        {
+            set_error( STATUS_INVALID_PARAMETER );
+            return;
+        }
+        memset( process->wine_cpu_id_from_host, 0, sizeof(process->wine_cpu_id_from_host) );
+        for (i = 0; i < cpu_override->cpu_count; ++i)
+        {
+            if (cpu_override->host_cpu_id[i] >= ARRAY_SIZE(process->wine_cpu_id_from_host))
+            {
+                set_error( STATUS_INVALID_PARAMETER );
+                return;
+            }
+            process->wine_cpu_id_from_host[cpu_override->host_cpu_id[i]] = i;
+        }
+        process->cpu_override = *cpu_override;
+    }
 
     if (!init_thread( current, req->reply_fd, req->wait_fd )) return;
 
@@ -1764,7 +1638,7 @@ DECL_HANDLER(init_first_thread)
     else
         set_thread_affinity( current, current->affinity );
 
-    set_thread_base_priority( current, current->base_priority );
+    set_thread_priority( current, process->priority, current->priority );
 
     debug_level = max( debug_level, req->debug_level );
 
@@ -1775,12 +1649,6 @@ DECL_HANDLER(init_first_thread)
     reply->server_start = server_start_time;
     set_reply_data( supported_machines,
                     min( supported_machines_count * sizeof(unsigned short), get_reply_max_size() ));
-
-    if ((fd = get_inproc_device_fd()) >= 0)
-    {
-        reply->inproc_device = get_process_id( process ) | 1;
-        send_client_fd( process, fd, reply->inproc_device );
-    }
 }
 
 /* initialize a new thread */
@@ -1801,10 +1669,10 @@ DECL_HANDLER(init_thread)
 
     init_thread_context( current );
     generate_debug_event( current, DbgCreateThreadStateChange, &req->entry );
-    set_thread_base_priority( current, current->base_priority );
+    apply_thread_priority( current, get_base_priority( current->process->priority, current->priority ));
     set_thread_affinity( current, current->affinity );
 
-    reply->suspend = (is_thread_suspended( current ) || current->context != NULL);
+    reply->suspend = (current->suspend || get_effective_proc_suspend( current ) || current->context != NULL);
 }
 
 /* terminate a thread */
@@ -1830,7 +1698,7 @@ DECL_HANDLER(open_thread)
     reply->handle = 0;
     if (thread)
     {
-        reply->handle = alloc_handle( current->process, thread, req->access, req->attributes );
+        reply->handle = alloc_handle_user_open( current->process, thread, req->access, req->attributes );
         release_object( thread );
     }
 }
@@ -1850,9 +1718,9 @@ DECL_HANDLER(get_thread_info)
         reply->teb            = thread->teb;
         reply->entry_point    = thread->entry_point;
         reply->exit_code      = (thread->state == TERMINATED) ? thread->exit_code : STATUS_PENDING;
-        reply->priority       = get_effective_thread_priority( thread );
-        reply->base_priority  = thread->base_priority;
+        reply->priority       = thread->priority;
         reply->affinity       = thread->affinity;
+        reply->last           = thread->process->running_threads == 1;
         reply->suspend_count  = thread->suspend;
         reply->desc_len       = thread->desc_len;
         reply->flags          = 0;
@@ -1860,10 +1728,6 @@ DECL_HANDLER(get_thread_info)
             reply->flags |= GET_THREAD_INFO_FLAG_DBG_HIDDEN;
         if (thread->state == TERMINATED)
             reply->flags |= GET_THREAD_INFO_FLAG_TERMINATED;
-        if (thread->process->running_threads == 1)
-            reply->flags |= GET_THREAD_INFO_FLAG_LAST;
-        if (thread->disable_boost)
-            reply->flags |= GET_THREAD_INFO_FLAG_DISABLE_BOOST;
 
         if (thread->desc && get_reply_max_size())
         {
@@ -1912,12 +1776,32 @@ DECL_HANDLER(suspend_thread)
 {
     struct thread *thread;
 
-    if ((thread = get_thread_from_handle( req->handle, THREAD_SUSPEND_RESUME )))
+    if (req->waited_handle)
     {
-        if (thread->state == TERMINATED) set_error( STATUS_ACCESS_DENIED );
-        else reply->count = suspend_thread( thread );
-        release_object( thread );
+        struct context *context;
+
+        if (!(context = (struct context *)get_handle_obj( current->process, req->waited_handle,
+                                                          0, &context_ops )))
+            return;
+        close_handle( current->process, req->waited_handle ); /* avoid extra server call */
+        set_error( context->status );
+        release_object( context );
+        return;
+    }
+
+    if (!(thread = get_thread_from_handle( req->handle, THREAD_SUSPEND_RESUME ))) return;
+
+    if (thread->state != RUNNING) set_error( STATUS_ACCESS_DENIED );
+    else
+    {
+        reply->count = suspend_thread( thread );
+        if (!get_error() && thread != current && thread->context && thread->context->status == STATUS_PENDING)
+        {
+            set_error( STATUS_PENDING );
+            reply->wait_handle = alloc_handle( current->process, thread->context, SYNCHRONIZE, 0 );
+        }
     }
+    release_object( thread );
 }
 
 /* resume a thread */
@@ -1989,7 +1873,7 @@ DECL_HANDLER(select)
         }
         ctx->status = STATUS_SUCCESS;
         current->suspend_cookie = req->cookie;
-        signal_sync( ctx->sync );
+        wake_up( &ctx->obj, 0 );
     }
 
     if (!req->cookie) goto invalid_param;
@@ -2013,7 +1897,7 @@ DECL_HANDLER(select)
             apc->result.create_thread.handle = handle;
             clear_error();  /* ignore errors from the above calls */
         }
-        signal_sync( apc->sync );
+        wake_up( &apc->obj, 0 );
         close_handle( current->process, req->prev_apc );
         release_object( apc );
     }
@@ -2036,7 +1920,7 @@ DECL_HANDLER(select)
         else
         {
             apc->executed = 1;
-            signal_sync( apc->sync );
+            wake_up( &apc->obj, 0 );
         }
         release_object( apc );
     }
@@ -2048,9 +1932,7 @@ DECL_HANDLER(select)
         {
             union apc_call *data;
             data_size_t size = sizeof(*data) + (ctx->regs[CTX_WOW].flags ? 2 : 1) * sizeof(struct context_data);
-            unsigned int flags = system_flags & ctx->regs[CTX_NATIVE].flags;
 
-            if (flags) set_thread_context( current, &ctx->regs[CTX_NATIVE], flags );
             size = min( size, get_reply_max_size() );
             if ((data = set_reply_data_size( size )))
             {
@@ -2083,20 +1965,7 @@ DECL_HANDLER(queue_apc)
     {
     case APC_NONE:
     case APC_USER:
-        if (req->reserve_handle &&
-            !(apc->reserve = reserve_obj_associate_apc( current->process, req->reserve_handle, &apc->obj )))
-        {
-            release_object( apc );
-            return;
-        }
         thread = get_thread_from_handle( req->handle, THREAD_SET_CONTEXT );
-        if (thread && call && call->user.flags & SERVER_USER_APC_SPECIAL && is_wow64_process( thread->process ))
-        {
-            release_object( apc );
-            release_object( thread );
-            set_error( STATUS_NOT_SUPPORTED );
-            return;
-        }
         break;
     case APC_VIRTUAL_ALLOC:
     case APC_VIRTUAL_ALLOC_EX:
@@ -2344,6 +2213,17 @@ DECL_HANDLER(set_thread_context)
     release_object( thread );
 }
 
+/* fetch a selector entry for a thread */
+DECL_HANDLER(get_selector_entry)
+{
+    struct thread *thread;
+    if ((thread = get_thread_from_handle( req->handle, THREAD_QUERY_INFORMATION )))
+    {
+        get_selector_entry( thread, req->entry, &reply->base, &reply->limit, &reply->flags );
+        release_object( thread );
+    }
+}
+
 /* Iterate thread list for process. Use global thread list to also
  * return terminated but not yet destroyed threads. */
 DECL_HANDLER(get_next_thread)
@@ -2380,11 +2260,15 @@ DECL_HANDLER(get_next_thread)
     while (ptr)
     {
         thread = LIST_ENTRY( ptr, struct thread, entry );
-        if (thread->process == process &&
-            (reply->handle = alloc_handle( current->process, thread, req->access, req->attributes )))
+        if (thread->process == process)
         {
-            release_object( process );
-            return;
+            reply->handle = alloc_handle_user_open( current->process, thread, req->access, req->attributes );
+            if (get_error() != STATUS_ACCESS_DENIED)
+            {
+                release_object( process );
+                return;
+            }
+            clear_error();
         }
         ptr = req->flags ? list_prev( &thread_list, &thread->entry )
                          : list_next( &thread_list, &thread->entry );
@@ -2392,17 +2276,3 @@ DECL_HANDLER(get_next_thread)
     set_error( STATUS_NO_MORE_ENTRIES );
     release_object( process );
 }
-
-
-/* Get the in-process synchronization fd for the current thread user APC alerts */
-DECL_HANDLER(get_inproc_alert_fd)
-{
-    int fd;
-
-    if ((fd = get_inproc_sync_fd( current->alert_sync )) < 0) set_error( STATUS_INVALID_PARAMETER );
-    else
-    {
-        reply->handle = get_thread_id( current ) | 1; /* arbitrary token */
-        send_client_fd( current->process, fd, reply->handle );
-    }
-}
