diff --git a/server/thread.h b/server/thread.h
index 77ea355483da5f18e33c14803ab1490fd7ebdbb3..d9b057311a7665b17cf5753f049cf5b8b6e3509c 100644
--- a/server/thread.h
+++ b/server/thread.h
@@ -50,15 +50,16 @@ struct inflight_fd
 struct thread
 {
     struct object          obj;           /* object header */
-    struct object         *sync;          /* sync object for wait/signal */
-    struct inproc_sync    *alert_sync;    /* inproc sync for user apc alerts */
     struct list            entry;         /* entry in system-wide thread list */
     struct list            proc_entry;    /* entry in per-process thread list */
     struct list            desktop_entry; /* entry in per-desktop thread list */
     struct process        *process;
     thread_id_t            id;            /* thread id */
     struct list            mutex_list;    /* list of currently owned mutexes */
-    struct list            d3dkmt_mutexes;/* list of currently owned d3dkmt mutexes */
+    int                    esync_fd;      /* esync file descriptor (signalled on exit) */
+    int                    esync_apc_fd;  /* esync apc fd (signalled when APCs are present) */
+    unsigned int           fsync_idx;
+    unsigned int           fsync_apc_idx;
     unsigned int           system_regs;   /* which system regs have been set */
     struct msg_queue      *queue;         /* message queue */
     struct thread_wait    *wait;          /* current wait condition if sleeping */
@@ -84,9 +85,7 @@ struct thread
     client_ptr_t           teb;           /* TEB address (in client address space) */
     client_ptr_t           entry_point;   /* entry point (in client address space) */
     affinity_t             affinity;      /* affinity mask */
-    int                    priority;      /* current thread priority */
-    int                    base_priority; /* base priority level (relative to process base priority class) */
-    int                    disable_boost; /* disable thread priority boost */
+    int                    priority;      /* priority level */
     int                    suspend;       /* suspend count */
     int                    dbg_hidden;    /* hidden from debugger */
     int                    bypass_proc_suspend; /* will still run if the process is suspended */
@@ -128,10 +127,7 @@ extern void thread_cancel_apc( struct thread *thread, struct object *owner, enum
 extern int thread_add_inflight_fd( struct thread *thread, int client, int server );
 extern int thread_get_inflight_fd( struct thread *thread, int client );
 extern struct token *thread_get_impersonation_token( struct thread *thread );
-extern int get_effective_thread_priority( struct thread *thread );
-extern unsigned int set_thread_priority( struct thread *thread, int priority );
-extern unsigned int set_thread_base_priority( struct thread *thread, int base_priority );
-extern void set_thread_disable_boost( struct thread *thread, int disable_boost );
+extern int set_thread_priority( struct thread *thread, int priority_class, int priority );
 extern int set_thread_affinity( struct thread *thread, affinity_t affinity );
 extern int suspend_thread( struct thread *thread );
 extern int resume_thread( struct thread *thread );
@@ -143,6 +139,8 @@ extern void init_thread_context( struct thread *thread );
 extern void get_thread_context( struct thread *thread, struct context_data *context, unsigned int flags );
 extern void set_thread_context( struct thread *thread, const struct context_data *context, unsigned int flags );
 extern int send_thread_signal( struct thread *thread, int sig );
+extern void get_selector_entry( struct thread *thread, int entry, unsigned int *base,
+                                unsigned int *limit, unsigned char *flags );
 
 extern unsigned int global_error;  /* global error code for when no thread is current */
 
