diff --git a/server/trace.c b/server/trace.c
index 8f0f622fb48e886d6c496cdbdda60a98b36421db..80621d579d38728f6ed5f1870fc0de2968c5cc86 100644
--- a/server/trace.c
+++ b/server/trace.c
@@ -1421,6 +1421,47 @@ static void dump_varargs_handle_infos( const char *prefix, data_size_t size )
     fputc( '}', stderr );
 }
 
+static void dump_varargs_cpu_topology_override( const char *prefix, data_size_t size )
+{
+    const struct cpu_topology_override *cpu_topology = cur_data;
+    unsigned int i;
+
+    if (size < sizeof(*cpu_topology))
+        return;
+
+    fprintf( stderr,"%s{", prefix );
+    for (i = 0; i < cpu_topology->cpu_count; ++i)
+    {
+        if (i) fputc( ',', stderr );
+        fprintf( stderr, "%u", cpu_topology->host_cpu_id[i] );
+    }
+    fputc( '}', stderr );
+    remove_data( size );
+}
+
+static void dump_varargs_directory_file_entries( const char *prefix, data_size_t size )
+{
+    const struct directory_file_entry *entry = cur_data;
+    const WCHAR *str;
+
+    fprintf( stderr, "%s{", prefix );
+    if (size)
+    {
+        if (size < sizeof(*entry) || size - sizeof(*entry) < entry->name_len)
+        {
+            fprintf( stderr, "***invalid***}" );
+            remove_data( size );
+            return;
+        }
+        str = (const WCHAR *)(entry + 1);
+        fprintf( stderr, "{name=L\"" );
+        dump_strW( str, entry->name_len, stderr, "\"\"" );
+        fprintf( stderr, "\"}" );
+        remove_data( size );
+    }
+    fputc( '}', stderr );
+}
+
 static void dump_varargs_tcp_connections( const char *prefix, data_size_t size )
 {
     static const char * const state_names[] = {
