diff --git a/tools/make_requests b/tools/make_requests
index 17e425537b55c7397a103c0940e6791ad2053bdc..7f1c9db3c95f57be6864d6309a5e0f937391f4a0 100755
--- a/tools/make_requests
+++ b/tools/make_requests
@@ -34,7 +34,6 @@ my %formats =
     "atom_t"        => [  4,   4,  "%04x" ],
     "process_id_t"  => [  4,   4,  "%04x" ],
     "thread_id_t"   => [  4,   4,  "%04x" ],
-    "d3dkmt_handle_t"=>[  4,   4,  "%08x" ],
     "unsigned __int64" => [ 8, 8,  "&uint64" ],
     "timeout_t"     => [  8,   8 ],
     "abstime_t"     => [  8,   8 ],
@@ -48,7 +47,7 @@ my %formats =
     "union debug_event_data"   => [ 160, 8 ],
     "struct filesystem_event"  => [ 12, 4 ],
     "struct generic_map"       => [ 16, 4 ],
-    "struct handle_info"       => [ 20, 4 ],
+    "struct handle_info"       => [ 32, 8 ],
     "union hw_input"           => [ 40, 8 ],
     "union irp_params"         => [ 32, 8 ],
     "struct luid"              => [  8, 4 ],
@@ -57,7 +56,7 @@ my %formats =
     "struct object_attributes" => [ 16, 4 ],
     "struct object_type_info"  => [ 44, 4 ],
     "struct obj_locator"       => [ 16, 8 ],
-    "struct pe_image_info"     => [ 96, 8 ],
+    "struct pe_image_info"     => [ 88, 8 ],
     "struct process_info"      => [ 40, 8 ],
     "struct property_data"     => [ 16, 8 ],
     "struct rawinput_device"   => [ 12, 4 ],
