diff --git a/tools/make_unicode b/tools/make_unicode
index f7ccbd8131d16731a56254b1f3ac8f714a3a8700..232b719947d5aa61a03b398bb03e3787fdc8555d 100755
--- a/tools/make_unicode
+++ b/tools/make_unicode
@@ -25,21 +25,21 @@ use Digest::SHA;
 use Encode;
 use Time::Local qw(timegm_modern);
 
-my $UNIVERSION = "17.0.0";
-my $CLDRVERSION = "48";
+my $UNIVERSION = "16.0.0";
+my $CLDRVERSION = "46";
 my $ISO639VERSION = "20230123";
-my $TZVERSION = "2025a";
+my $TZVERSION = "2024a";
 
 my %data_files =
 (
  ucd       => { url  => "https://www.unicode.org/Public/$UNIVERSION/ucd/UCD.zip", name => "UCD-$UNIVERSION.zip",
-                sha  => "2066d1909b2ea93916ce092da1c0ee4808ea3ef8407c94b4f14f5b7eb263d28e" },
+                sha  => "c86dd81f2b14a43b0cc064aa5f89aa7241386801e35c59c7984e579832634eb2" },
  unihan    => { url  => "https://www.unicode.org/Public/$UNIVERSION/ucd/Unihan.zip", name => "Unihan-$UNIVERSION.zip",
-                sha  => "f7a48b2b545acfaa77b2d607ae28747404ce02baefee16396c5d2d7a8ef34b5e" },
- idna      => { url  => "https://www.unicode.org/Public/$UNIVERSION/idna/IdnaMappingTable.txt", name => "IdnaMappingTable-$UNIVERSION.txt",
-                sha  => "87f05505dc026fdb2bff16132bdc68a8014675836882a9a2b1844540ad3be382" },
+                sha  => "b8f000df69de7828d21326a2ffea462b04bc7560022989f7cc704f10521ef3e0" },
+ idna      => { url  => "https://www.unicode.org/Public/idna/$UNIVERSION/IdnaMappingTable.txt", name => "IdnaMappingTable-$UNIVERSION.txt",
+                sha  => "6db2ef4ed35f3b3de74ebc2e00404a9607f76d499f576b8d4043cf14f1ed175c" },
  cldr      => { url  => "https://github.com/unicode-org/cldr/archive/refs/tags/release-$CLDRVERSION.zip",
-                sha  => "fa6d331e52d59771009c424bd1c9df311b6b2bf0bf2f6c1ddf6c952c32da4702" },
+                sha  => "716d0fc10a6469f7b808e55671ab98640c387e43bae152382e725462e366ccb2" },
  cldr33    => { url  => "https://www.unicode.org/Public/cldr/33/cldr-common-33.0.zip",
                 sha  => "fa3490082c086d21257153609642f54fcf788fcfda4966fe67f3f6daca0d58b9" },
  sorting   => { url  => "https://download.microsoft.com/download/C/F/7/CF713A5E-9FBC-4FD6-9246-275F65C0E498/Windows 10 Sorting Weight Table.txt",
@@ -55,11 +55,7 @@ my %data_files =
  jis0212   => { url  => "https://www.unicode.org/Public/MAPPINGS/OBSOLETE/EASTASIA/JIS/JIS0212.TXT",
                 sha  => "477820bb3055bbcc90880d788cd95607d221dc94457bae249231adecf13c12e6" },
  tzdata    => { url  => "https://data.iana.org/time-zones/releases/tzdata$TZVERSION.tar.gz",
-                sha  => "4d5fcbc72c7c450ebfe0b659bd0f1c02fbf52fd7f517a9ea13fe71c21eb5f0d0" },
- glyphs    => { url  => "https://github.com/adobe-type-tools/agl-aglfn/raw/4036a9ca80a62f64f9de4f7321a9a045ad0ecfd6/glyphlist.txt",
-                sha  => "a3b2f61ced9f3644cc0d4ecde5c59df34ca286c689d9484a43a710a81c466789" },
- psfonts   => { url  => "https://mirrors.ctan.org/fonts/adobe/afm/Adobe-Core35_AFMs-229.tar.gz",
-                sha  => "034ed3025cff4c48bbefae90f9eff7dc6237dd384443c594c017b571e49d0649" },
+                sha  => "0d0434459acbd2059a7a8da1f3304a84a86591f6ed69c6248fffa502b6edffe3" },
 );
 
 
@@ -71,6 +67,7 @@ my $MAX_CHAR = 0x10ffff;
 
 my $nlskey = "-SYSTEM\\-CurrentControlSet\\-Control\\-Nls";
 my $zonekey = "-Software\\-Microsoft\\-Windows NT\\-CurrentVersion\\Time Zones";
+my $winezonekey = "-Software\\-Wine\\Time Zones";
 
 my @allfiles =
 (
@@ -290,7 +287,6 @@ my %break_types =
     "AS"  => 0x002e,
     "VF"  => 0x002f,
     "VI"  => 0x0030,
-    "HH"  => 0x0031,
 );
 
 my %vertical_types =
@@ -563,8 +559,6 @@ my @locales =
  { name => "bo-Tibt", alias => "bo" },
  { name => "bo-Tibt-CN", alias => "bo-CN" },
  { name => "bo-Tibt-IN", alias => "bo-IN" },
- { name => "bqi" },
- { name => "bqi-IR" },
  { name => "br", lcid => 0x0000007e, oemcp => 850, ebcdiccp => 20297 },
  { name => "br-FR", lcid => 0x0000047e },
  { name => "br-Latn", alias => "br" },
@@ -576,8 +570,6 @@ my @locales =
  { name => "bs-Cyrl-BA", lcid => 0x0000201a },
  { name => "bs-Latn", lcid => 0x0000681a },
  { name => "bs-Latn-BA", lcid => 0x0000141a },
- { name => "bua" },
- { name => "bua-RU" },
  { name => "byn", sopentypelang => "BIL" },
  { name => "byn-ER" },
  { name => "ca", lcid => 0x00000003, oemcp => 850 },
@@ -586,8 +578,6 @@ my @locales =
  { name => "ca-ES-valencia", lcid => 0x00000803, file => "ca_ES_VALENCIA", sabbrevlangname => "VAL" },
  { name => "ca-FR", maccp => 65001 },
  { name => "ca-IT", maccp => 65001 },
- { name => "cop" },
- { name => "cop-EG" },
  { name => "ccp" },
  { name => "ccp-BD", alias => "ccp-Cakm-BD" },
  { name => "ccp-Cakm", file => "ccp" },
@@ -606,9 +596,9 @@ my @locales =
  { name => "chr-Cher", lcid => 0x00007c5c, file => "chr" },
  { name => "chr-Cher-US", lcid => 0x0000045c, file => "chr_US" },
  { name => "chr-US", alias => "chr-Cher-US" },
- { name => "ckb" },
- { name => "ckb-IQ" },
- { name => "ckb-IR" },
+ { name => "ckb", alias => "ku" },
+ { name => "ckb-IQ", alias => "ku-Arab-IQ" },
+ { name => "ckb-IR", alias => "ku-Arab-IR" },
  { name => "co", lcid => 0x00000083, oemcp => 850, ebcdiccp => 20297 },
  { name => "co-FR", lcid => 0x00000483 },
  { name => "co-Latn", alias => "co" },
@@ -688,7 +678,6 @@ my @locales =
  { name => "en-DG", oemcp => 850 },
  { name => "en-DK", oemcp => 65001 },
  { name => "en-DM", oemcp => 850 },
- { name => "en-EE", oemcp => 850 },
  { name => "en-ER", oemcp => 850 },
  { name => "en-FI", oemcp => 65001 },
  { name => "en-FJ", oemcp => 850 },
@@ -696,7 +685,6 @@ my @locales =
  { name => "en-FM", oemcp => 850 },
  { name => "en-GB", lcid => 0x00000809, oemcp => 850, ebcdiccp => 20285, sabbrevlangname => "ENG" },
  { name => "en-GD", oemcp => 850 },
- { name => "en-GE", oemcp => 850 },
  { name => "en-GG", oemcp => 850 },
  { name => "en-GH", oemcp => 850 },
  { name => "en-GI", oemcp => 850 },
@@ -719,8 +707,6 @@ my @locales =
  { name => "en-LC", oemcp => 850 },
  { name => "en-LR", oemcp => 850 },
  { name => "en-LS", oemcp => 850 },
- { name => "en-LT", oemcp => 850 },
- { name => "en-LV", oemcp => 850 },
  { name => "en-MG", oemcp => 850 },
  { name => "en-MH", oemcp => 850 },
  { name => "en-MO", oemcp => 850 },
@@ -761,7 +747,6 @@ my @locales =
  { name => "en-TT", lcid => 0x00002c09, oemcp => 850, sabbrevlangname => "ENT" },
  { name => "en-TV", oemcp => 850 },
  { name => "en-TZ", oemcp => 850 },
- { name => "en-UA", oemcp => 850 },
  { name => "en-UG", oemcp => 850 },
  { name => "en-UM", oemcp => 850 },
  { name => "en-US", lcid => 0x00000409 },
@@ -951,8 +936,6 @@ my @locales =
  { name => "hr-HR", lcid => 0x0000041a },
  { name => "hsb", lcid => 0x0000002e, oemcp => 850, ebcdiccp => 870, sopentypelang => "USB" },
  { name => "hsb-DE", lcid => 0x0000042e },
- { name => "ht" },
- { name => "ht-HT" },
  { name => "hu", lcid => 0x0000000e, oemcp => 852, group => 2 },
  { name => "hu-HU", lcid => 0x0000040e },
  { name => "hu-HU_technl", lcid => 0x0001040e, alias => "hu-HU" },
@@ -1017,8 +1000,6 @@ my @locales =
  { name => "kde-TZ" },
  { name => "kea" },
  { name => "kea-CV" },
- { name => "kek" },
- { name => "kek-GT" },
  { name => "kgp" },
  { name => "kgp-BR" },
  { name => "khq" },
@@ -1066,14 +1047,10 @@ my @locales =
  { name => "ksf-CM" },
  { name => "ksh", sopentypelang => "KSH0" },
  { name => "ksh-DE" },
- { name => "ku", lcid => 0x00000092, slist => "\x{061b}", sortlocale => "ku-Arab-IQ", oemcp => 720 },
- { name => "ku-Arab", lcid => 0x00007c92, group => 13 },
- { name => "ku-Arab-IQ", lcid => 0x00000492 },
- { name => "ku-Arab-IR", oemcp => 65001 },
- { name => "ku-Latn" },
- { name => "ku-Latn-IQ" },
- { name => "ku-Latn-SY" },
- { name => "ku-Latn-TR" },
+ { name => "ku", lcid => 0x00000092, file => "ckb", slist => "\x{061b}", sortlocale => "ku-Arab-IQ", oemcp => 720 },
+ { name => "ku-Arab", lcid => 0x00007c92, file => "ckb", group => 13 },
+ { name => "ku-Arab-IQ", lcid => 0x00000492, file => "ckb_IQ" },
+ { name => "ku-Arab-IR", file => "ckb_IR", oemcp => 65001 },
  { name => "kw" },
  { name => "kw-GB" },
  { name => "ky", lcid => 0x00000040, oemcp => 866, group => 5, sabbrevlangname => "KYR" },
@@ -1115,8 +1092,6 @@ my @locales =
  { name => "luy-KE" },
  { name => "lv", lcid => 0x00000026, oemcp => 775, group => 3, sabbrevlangname => "LVI", sopentypelang => "LVI" },
  { name => "lv-LV", lcid => 0x00000426 },
- { name => "lzz" },
- { name => "lzz-TR" },
  { name => "mai" },
  { name => "mai-IN" },
  { name => "mas" },
@@ -1174,9 +1149,6 @@ my @locales =
  { name => "mt-MT", lcid => 0x0000043a },
  { name => "mua" },
  { name => "mua-CM" },
- { name => "mww" },
- { name => "mww-Hmnp" },
- { name => "mww-Hmnp-US" },
  { name => "my", lcid => 0x00000055, sopentypelang => "BRM" },
  { name => "my-MM", lcid => 0x00000455 },
  { name => "mzn" },
@@ -1224,9 +1196,6 @@ my @locales =
  { name => "oc-FR", lcid => 0x00000482 },
  { name => "oc-Latn", alias => "oc" },
  { name => "oc-Latn-FR", alias => "oc-FR" },
- { name => "oka" },
- { name => "oka-CA" },
- { name => "oka-US" },
  { name => "om", lcid => 0x00000072, sopentypelang => "ORO" },
  { name => "om-ET", lcid => 0x00000472 },
  { name => "om-KE" },
@@ -1247,13 +1216,8 @@ my @locales =
  { name => "pcm-NG", alias => "pcm-Latn-NG" },
  { name => "pcm-Latn", file => "pcm" },
  { name => "pcm-Latn-NG", file => "pcm_NG" },
- { name => "pi" },
- { name => "pi-Latn" },
- { name => "pi-Latn-GB" },
  { name => "pl", lcid => 0x00000015, oemcp => 852, ebcdiccp => 20880, group => 2, sabbrevlangname => "PLK", sopentypelang => "PLK" },
  { name => "pl-PL", lcid => 0x00000415 },
- { name => "pms" },
- { name => "pms-IT" },
  { name => "prg" },
  { name => "prg-001", file => "prg" },
  { name => "prg-PL" },
@@ -1355,8 +1319,6 @@ my @locales =
  { name => "ses-ML" },
  { name => "sg", sopentypelang => "SGO" },
  { name => "sg-CF" },
- { name => "sgs" },
- { name => "sgs-LT" },
  { name => "shi" },
  { name => "shi-Latn" },
  { name => "shi-Latn-MA" },
@@ -1427,11 +1389,6 @@ my @locales =
  { name => "su" },
  { name => "su-Latn" },
  { name => "su-Latn-ID" },
- { name => "suz" },
- { name => "suz-Deva" },
- { name => "suz-Deva-NP" },
- { name => "suz-Sunu" },
- { name => "suz-Sunu-NP" },
  { name => "sv", lcid => 0x0000001d, oemcp => 850, ebcdiccp => 20278, sabbrevlangname => "SVE", sopentypelang => "SVE" },
  { name => "sv-AX" },
  { name => "sv-FI", lcid => 0x0000081d, sabbrevlangname => "SVF" },
@@ -1568,7 +1525,6 @@ my @locales =
  { name => "yue-Hant" },
  { name => "yue-Hant-CN" },
  { name => "yue-Hant-HK" },
- { name => "yue-Hant-MO" },
  { name => "zgh" },
  { name => "zgh-MA", alias => "zgh-Tfng-MA" },
  { name => "zgh-Tfng", file => "zgh" },
@@ -2017,11 +1973,7 @@ sub download_data_files()
         {
             print "Fetching $url...\n";
             system "mkdir", "-p", $cache;
-            if (system "wget", "-q", "-O", $filename, $url)
-            {
-                unlink $filename;
-                die "cannot fetch $url";
-            }
+            !system "wget", "-q", "-O", $filename, $url or die "cannot fetch $url";
         }
         my $sha = Digest::SHA->new( "sha256" )->addfile( $filename )->hexdigest;
         die "invalid checksum $sha for $filename" unless $sha eq $data->{sha};
@@ -3047,22 +2999,6 @@ my %scripts =
     "Dives_Akuru"            => 154,
     "Khitan_Small_Script"    => 155,
     "Yezidi"                 => 156,
-    # Win11 24H2
-    "Cypro_Minoan"           => 157,
-    "Kawi"                   => 158,
-    "Nag_Mundari"            => 159,
-    "Old_Uyghur"             => 160,
-    "Tangsa"                 => 161,
-    "Toto"                   => 162,
-    "Vithkuqi"               => 163,
-    # Win11 25H2
-    "Garay"                  => 164,
-    "Gurung_Khema"           => 165,
-    "Kirat_Rai"              => 166,
-    "Ol_Onal"                => 167,
-    "Sunuwar"                => 168,
-    "Todhri"                 => 169,
-    "Tulu_Tigalari"          => 170,
 );
 
 ################################################################
@@ -4802,9 +4738,9 @@ sub build_locale_data()
         my $sopentypelang = sprintf "%-4s", locale_entry( $loc, "sopentypelang", uc $siso639langname2 );
         my $sabbrevlangname = defined $loc->{lcid} ? locale_entry( $loc, "sabbrevlangname", uc $siso639langname2 ) : "ZZZ";
         my $siso3166ctryname2 = $geo->{iso3} || $geo->{uncode};
-        my $senglanguage = loc_query( $lcnames{en}, "/ldml/localeDisplayNames/languages/language[\@type='$language' and not(\@alt) and not(\@menu)]" ) || "";
+        my $senglanguage = loc_query( $lcnames{en}, "/ldml/localeDisplayNames/languages/language[\@type='$language' and not(\@alt)]" ) || "";
         my $sengcountry = loc_query( $lcnames{en}, "/ldml/localeDisplayNames/territories/territory[\@type='$territory' and not(\@alt)]" ) || "";
-        my $snativelangname = loc_query( $loc, "/ldml/localeDisplayNames/languages/language[\@type='$language' and not(\@alt) and not(\@menu)]" );
+        my $snativelangname = loc_query( $loc, "/ldml/localeDisplayNames/languages/language[\@type='$language' and not(\@alt)]" );
         my $snativectryname = loc_query( $loc, "/ldml/localeDisplayNames/territories/territory[\@type='$territory' and not(\@alt)]" );
         $sengcountry =~ s/South Korea/Korea/;
         $sengcountry =~ s/T\xfcrkiye/Turkey/;
@@ -4917,7 +4853,7 @@ sub build_locale_data()
         {
             next unless defined $caltypes{$c};
             $icalendartype .= chr($caltypes{$c});
-            $scalnames[$caltypes{$c} - 1] = loc_query( $loc, "/ldml/localeDisplayNames/types/type[\@key='calendar' and \@type='$c' and not(\@scope)]" );
+            $scalnames[$caltypes{$c} - 1] = loc_query( $loc, "/ldml/localeDisplayNames/types/type[\@key='calendar' and \@type='$c']" );
         }
 
         # date/time formats
@@ -5215,7 +5151,7 @@ sub build_locale_data()
                 push @sabbrevmonthname, $abbrev || $name || "";
             }
 
-            $scalname ||= loc_query( $loc, "/ldml/localeDisplayNames/types/type[\@key='calendar' and \@type='$type' and not(\@scope)]" );
+            $scalname ||= loc_query( $loc, "/ldml/localeDisplayNames/types/type[\@key='calendar' and \@type='$type']" );
             if (defined $cal->{eras})
             {
                 my @eras;
@@ -5501,6 +5437,7 @@ sub load_windows_zones()
 {
     my $current_name;
     my %names;
+    my %mapping;
     my $base = "cldr-release-$CLDRVERSION";
     my $INPUT = open_data_file( "cldr", "$base/common/supplemental/windowsZones.xml" );
     while (<$INPUT>)
@@ -5513,9 +5450,16 @@ sub load_windows_zones()
         {
             $names{$1} = [ $current_name, $2 ];
         }
+        if (/<mapZone other="(.*)" territory=(.*) type="(.*)"\/>/)
+        {
+            foreach my $name (split ' ', $3)
+            {
+                $mapping{$name} = $1;
+            }
+        }
     }
     close $INPUT;
-    return %names;
+    return (\%names, \%mapping);
 }
 
 
@@ -5533,7 +5477,7 @@ sub parse_transition_date($$@)
                    Jul => 7, Aug => 8, Sep => 9, Oct => 10, Nov => 11, Dec => 12 );
     my %days = ( Sun => 0, Mon => 1, Tue => 2, Wed => 3, Thu => 4, Fri => 5, Sat => 6 );
 
-    my $mon = $in ? $months{substr($in,0,3)} : 1;
+    my $mon = $in ? $months{$in} : 1;
     my ($week, $dow, $flag, $time, $sec);
     my $first = month_first_dow( $year, $mon );
 
@@ -5638,7 +5582,7 @@ sub dump_timezones($@)
     my $FIRST_YEAR = 2000;
     my $LAST_YEAR = 2030;
 
-    my %names = load_windows_zones();
+    my ($names, $mapping) = load_windows_zones();
     my %zones;
     my %rules;
     my %links;
@@ -5694,9 +5638,10 @@ sub dump_timezones($@)
         close $FILE;
     }
 
-    foreach my $name (sort { uc($a) cmp uc($b) } keys %names)
+    foreach my $name (sort { uc($a) cmp uc($b) } keys %{$names})
     {
-        my ($display, $zone) = @{$names{$name}};
+        my ($display, $zone) = @{%{$names}{$name}};
+
         $zone = $links{$zone} if defined $links{$zone};
 
         # build list of transitions
@@ -5865,6 +5810,12 @@ sub dump_timezones($@)
         add_registry_dword_value( $zonekey, "$name\\Dynamic DST", "LastEntry", $last_year );
     }
 
+    foreach my $map (sort { uc($a) cmp uc($b) } keys %{$mapping})
+    {
+        my $name = %{$mapping}{$map};
+        add_registry_string_value( $winezonekey, "TZ Mapping", $map, $name );
+    }
+
     print OUTPUT "}\n";
     close OUTPUT;
     save_file($filename);
@@ -5915,294 +5866,6 @@ sub dump_registry_script($%)
 }
 
 
-sub is_win_ansi($)
-{
-    my $c = shift;
-    return 1 if $c >= 0x0020 && $c <= 0x007e;
-    return 1 if $c >= 0x00a0 && $c <= 0x00ff;
-    return 1 if $c >= 0x0152 && $c <= 0x0153;
-    return 1 if $c >= 0x0160 && $c <= 0x0161;
-    return 1 if $c >= 0x017d && $c <= 0x017e;
-    return 1 if $c >= 0x2013 && $c <= 0x2014;
-    return 1 if $c >= 0x2018 && $c <= 0x201a;
-    return 1 if $c >= 0x201c && $c <= 0x201e;
-    return 1 if $c >= 0x2020 && $c <= 0x2022;
-    return 1 if $c >= 0x2039 && $c <= 0x203a;
-    return 1 if $c == 0x0178;
-    return 1 if $c == 0x0192;
-    return 1 if $c == 0x02c6;
-    return 1 if $c == 0x02c9;
-    return 1 if $c == 0x02dc;
-    return 1 if $c == 0x03bc;
-    return 1 if $c == 0x2026;
-    return 1 if $c == 0x2030;
-    return 1 if $c == 0x20ac;
-    return 1 if $c == 0x2122;
-    return 1 if $c == 0x2219;
-    return 0;
-}
-
-################################################################
-# build the PostScript font metrics
-sub dump_psfonts($)
-{
-    my $dir = shift;
-
-    my @psfonts =
-    (
-     "Courier",
-     "Courier-Bold",
-     "Courier-BoldOblique",
-     "Courier-Oblique",
-     "Helvetica",
-     "Helvetica-Bold",
-     "Helvetica-BoldOblique",
-     "Helvetica-Narrow",
-     "Helvetica-NarrowBold",
-     "Helvetica-NarrowBoldOblique",
-     "Helvetica-NarrowOblique",
-     "Helvetica-Oblique",
-     "ITCAvantGarde-Book",
-     "ITCAvantGarde-BookOblique",
-     "ITCAvantGarde-Demi",
-     "ITCAvantGarde-DemiOblique",
-     "ITCBookman-Demi",
-     "ITCBookman-DemiItalic",
-     "ITCBookman-Light",
-     "ITCBookman-LightItalic",
-     "ITCZapfChancery-MediumItalic",
-     "NewCenturySchlbk-Bold",
-     "NewCenturySchlbk-BoldItalic",
-     "NewCenturySchlbk-Italic",
-     "NewCenturySchlbk-Roman",
-     "Palatino-Bold",
-     "Palatino-BoldItalic",
-     "Palatino-Italic",
-     "Palatino-Roman",
-     "Symbol",
-     "Times-Bold",
-     "Times-BoldItalic",
-     "Times-Italic",
-     "Times-Roman",
-     "ZapfDingbats",
-    );
-
-    my %fonts;
-    my %glyphs;
-    my %glyph_values;
-
-    my $FILE = open_data_file( "glyphs" );
-    while (<$FILE>)
-    {
-        chomp;
-        s/\r$//;
-        next if /^$/;
-        next if /^#/;
-        if (/([A-Za-z0-9_]+);([0-9A-F]+)/)
-        {
-            $glyph_values{$2} ||= $1;
-            $glyphs{$1} ||= hex $2;
-            next;
-        }
-        die "unrecognized line $_";
-    }
-
-    my $filename = "$dir/glyphs.c";
-    printf "Building %s\n", $filename;
-    my $count = scalar keys %glyphs;
-    open OUTPUT, ">$filename.new" or die "Cannot create $filename";
-    print OUTPUT "/* Adobe Glyph List, generated from $current_data_file */\n";
-    print OUTPUT "/* DO NOT EDIT!! */\n\n";
-    print OUTPUT "#include \"psdrv.h\"\n\n";
-    print OUTPUT "const INT PSDRV_AGLbyNameSize = $count;\n\n";
-    print OUTPUT "const UNICODEGLYPH PSDRV_AGLbyName[$count] =\n{\n";
-    map { printf OUTPUT "    { 0x%04x, \"$_\" },\n", $glyphs{$_}; } sort keys %glyphs;
-    print OUTPUT "};\n\n";
-
-    $count = scalar keys %glyph_values;
-    print OUTPUT "const INT PSDRV_AGLbyUVSize = $count;\n\n";
-    print OUTPUT "const UNICODEGLYPH PSDRV_AGLbyUV[$count] =\n{\n";
-    map { printf OUTPUT "    { 0x%04x, \"%s\" },\n", hex $_, $glyph_values{$_}; } sort keys %glyph_values;
-    print OUTPUT "};\n";
-    close OUTPUT;
-    save_file($filename);
-
-    foreach my $file (@psfonts)
-    {
-        my $FILE = open_data_file( "psfonts", "Adobe-Core35_AFMs-229/$file.afm" );
-        my %font;
-        while (<$FILE>)
-        {
-            chomp;
-            last if /^StartFontMetrics\s/;
-            die "unrecognized line $_";
-        }
-        while (<$FILE>)
-        {
-            chomp;
-            last if /^StartCharMetrics\s/;
-            next if /^Comment\s/;
-            if (/^([A-Za-z]+)\s+(.*)/)
-            {
-                $font{$1} = $2;
-                next;
-            }
-            die "unrecognized line $_";
-        }
-        $font{chars} = [ ];
-        my $ascender = 0;
-        my $descender = 0;
-        while (<$FILE>)
-        {
-            chomp;
-            last if /^EndCharMetrics$/;
-            my @fields = split /\s*;\s*/;
-            my %char;
-            foreach my $f (@fields)
-            {
-                if ($f =~ /^C\s+(-?\d+)/)
-                {
-                    $char{C} = $1;
-                }
-                elsif ($f =~ /^WX\s+(\d+)/)
-                {
-                    $char{WX} = $1;
-                }
-                elsif ($f =~ /^N\s+([A-Za-z0-9_]+)/)
-                {
-                    $char{N} = $1;
-                    if ($font{EncodingScheme} eq "FontSpecific")
-                    {
-                        if ($char{C} == -1)
-                        {
-                            %char = ();
-                            last;
-                        }
-                        $char{UV} = 0xf000 + $char{C};
-                    }
-                    elsif (defined $glyphs{$1})
-                    {
-                        $char{UV} = $glyphs{$1};
-                    }
-                    else
-                    {
-                        %char = ();
-                        last;
-                    }
-                }
-                elsif ($f =~ /^B\s+(.*)/)
-                {
-                    next if $font{EncodingScheme} eq "FontSpecific";
-                    next unless is_win_ansi( $char{C} );
-                    my @bbox = split /\s+/, $1;
-                    $ascender = $bbox[3] if $bbox[3] > $ascender;
-                    $descender = $bbox[1] if $bbox[1] < $descender;
-                }
-                elsif ($f =~ /^L\s+/)
-                {
-                    next;
-                }
-                else { die "unrecognized field $f"; }
-            }
-            push @{$font{chars}}, \%char if %char;
-        }
-        close $FILE;
-
-        $font{FamilyName} .= " Narrow" if $font{FullName} =~ / Narrow/;
-        my @bbox = split /\s+/, $font{FontBBox};
-        $ascender ||= $bbox[3];
-        $descender ||= $bbox[1];
-        my $linegap = 1150 - ($ascender - $descender);
-        $font{FontBBox} = \@bbox;
-        $font{usUnitsPerEm} = 1000;
-        $font{sAscender} = $ascender;
-        $font{sDescender} = $descender;
-        $font{sLineGap} = $linegap > 0 ? $linegap : 0;
-        $font{usWinAscent} = $ascender;
-        $font{usWinDescent} = -$descender;
-
-        if ($font{EncodingScheme} eq "FontSpecific")
-        {
-            my $width = 0;
-            foreach my $m (@{$font{chars}})
-            {
-                $width += $m->{WX};
-            }
-            $font{sAvgCharWidth} = $width / scalar @{$font{chars}};
-        }
-        else
-        {
-            my @weights;
-            $weights[ord(' ')] = 166;
-            @weights[ord('a')..ord('z')] = ( 64, 14, 27,  35, 100, 20, 14, 42, 63, 3, 6, 35, 20, 56, 56, 17, 4, 49, 56, 71, 31, 10, 18, 3, 18, 2 );
-            my $width = 0;
-            foreach my $m (@{$font{chars}})
-            {
-                next unless defined $weights[$m->{UV}];
-                $width += $weights[$m->{UV}] * $m->{WX};
-            }
-            $font{sAvgCharWidth} = ($width + 500) / 1000;
-        }
-        (my $varname = $font{FontName}) =~ s/-/_/g;
-        $fonts{$varname} = \%font;
-    }
-
-    $filename = "$dir/corefonts.c";
-    printf "Building %s\n", $filename;
-    open OUTPUT, ">$filename.new" or die "Cannot create $filename";
-    print OUTPUT "/* Adobe Core Fonts, generated from $data_files{psfonts}->{url} */\n";
-    print OUTPUT "/* DO NOT EDIT!! */\n\n";
-    print OUTPUT "#include \"psdrv.h\"\n\n";
-
-    foreach my $name (sort keys %fonts)
-    {
-        my $count = scalar @{$fonts{$name}->{chars}};
-        printf OUTPUT "static const AFMMETRICS metrics_%s[%u] =\n{\n", $name, $count;
-        foreach my $m (sort { $a->{UV} <=> $b->{UV} } @{$fonts{$name}->{chars}})
-        {
-            printf OUTPUT "    { 0x%04x, %4u },\n", $m->{UV}, $m->{WX};
-        }
-        print OUTPUT  "};\n\n";
-    }
-
-    foreach my $name (sort keys %fonts)
-    {
-        my $count = scalar @{$fonts{$name}->{chars}};
-        my %weights = ( Bold => "FW_BOLD",
-                        Book => "FW_NORMAL",
-                        Demi => "FW_BOLD",
-                        Light => "FW_NORMAL",
-                        Roman => "FW_NORMAL",
-                        Medium => "FW_NORMAL",
-            );
-        printf OUTPUT "static const AFM font_%s =\n{\n", $name;
-        printf OUTPUT "    .FontName = \"%s\",\n", $fonts{$name}->{FontName};
-        printf OUTPUT "    .FamilyName = L\"%s\",\n", $fonts{$name}->{FamilyName};
-        printf OUTPUT "    .EncodingScheme = L\"%s\",\n", $fonts{$name}->{EncodingScheme};
-        printf OUTPUT "    .Weight = %s,\n", $weights{$fonts{$name}->{Weight}};
-        printf OUTPUT "    .ItalicAngle = %s,\n", $fonts{$name}->{ItalicAngle};
-        printf OUTPUT "    .IsFixedPitch = %s,\n", uc $fonts{$name}->{IsFixedPitch};
-        printf OUTPUT "    .FontBBox = { %d, %d, %d, %d },\n", @{$fonts{$name}->{FontBBox}}[0..3];
-        printf OUTPUT "    .WinMetrics.usUnitsPerEm = %d,\n", $fonts{$name}->{usUnitsPerEm};
-        printf OUTPUT "    .WinMetrics.sAscender = %d,\n", $fonts{$name}->{sAscender};
-        printf OUTPUT "    .WinMetrics.sDescender = %d,\n", $fonts{$name}->{sDescender};
-        printf OUTPUT "    .WinMetrics.sLineGap = %d,\n", $fonts{$name}->{sLineGap};
-        printf OUTPUT "    .WinMetrics.sAvgCharWidth = %d,\n", $fonts{$name}->{sAvgCharWidth};
-        printf OUTPUT "    .WinMetrics.usWinAscent = %d,\n", $fonts{$name}->{usWinAscent};
-        printf OUTPUT "    .WinMetrics.usWinDescent = %d,\n", $fonts{$name}->{usWinDescent};
-        printf OUTPUT "    .NumofMetrics = %s,\n", $count;
-        printf OUTPUT "    .Metrics = metrics_%s\n", $name;
-        print OUTPUT  "};\n\n";
-    }
-
-    print OUTPUT "const AFM * const PSDRV_BuiltinAFMs[] =\n{\n";
-    map { print OUTPUT  "    &font_$_,\n"; } sort keys %fonts;
-    print OUTPUT  "    NULL\n};\n";
-    close OUTPUT;
-    save_file($filename);
-}
-
-
 ################################################################
 # save a file if modified
 sub save_file($)
@@ -6240,7 +5903,6 @@ dump_scripts( "dlls/dwrite/scripts" );
 dump_indic( "dlls/gdi32/uniscribe/indicsyllable.c" );
 dump_vertical( "dlls/win32u/vertical.c", 1 );
 dump_vertical( "dlls/wineps.drv/vertical.c", 0 );
-dump_psfonts( "dlls/wineps.drv" );
 dump_intl_nls("nls/l_intl.nls");
 dump_norm_table( "nls/normnfc.nls" );
 dump_norm_table( "nls/normnfd.nls" );
