diff --git a/tools/tools.h b/tools/tools.h
index be569f801d5d88124e5c3e54911e738f05f87447..b840a85b2e53eb841a65074966d7f2a5d7233927 100644
--- a/tools/tools.h
+++ b/tools/tools.h
@@ -27,7 +27,6 @@
 
 #include <limits.h>
 #include <stdarg.h>
-#include <stdbool.h>
 #include <stdio.h>
 #include <stdlib.h>
 #include <string.h>
@@ -40,9 +39,6 @@
 #ifdef HAVE_SYS_SYSCTL_H
 # include <sys/sysctl.h>
 #endif
-#ifdef __APPLE__
-# include <mach-o/dyld.h>
-#endif
 
 #ifdef _WIN32
 # include <direct.h>
@@ -52,9 +48,6 @@
 # ifndef S_ISREG
 #  define S_ISREG(mod) (((mod) & _S_IFMT) == _S_IFREG)
 # endif
-# ifndef S_ISDIR
-#  define S_ISDIR(mod) (((mod) & _S_IFMT) == _S_IFDIR)
-# endif
 # ifdef _MSC_VER
 #  define popen _popen
 #  define pclose _pclose
@@ -115,20 +108,6 @@ struct target
     } platform;
 };
 
-static void fatal_perror( const char *msg, ... ) __attribute__ ((__format__ (__printf__, 1, 2), noreturn));
-static inline void fatal_perror( const char *msg, ... )
-{
-    va_list valist;
-
-    va_start( valist, msg );
-    fprintf( stderr, "error: " );
-    vfprintf( stderr, msg, valist );
-    perror( " " );
-    va_end( valist );
-    exit(1);
-}
-
-
 static inline void *xmalloc( size_t size )
 {
     void *res = malloc( size ? size : 1 );
@@ -158,7 +137,7 @@ static inline char *xstrdup( const char *str )
     return strcpy( xmalloc( strlen(str)+1 ), str );
 }
 
-static inline bool strendswith( const char *str, const char *end )
+static inline int strendswith( const char *str, const char *end )
 {
     int l = strlen( str );
     int m = strlen( end );
@@ -185,35 +164,6 @@ static inline char *strmake( const char* fmt, ... )
     }
 }
 
-/* generic array functions */
-
-struct array
-{
-    size_t count;   /* count of array entries */
-    size_t size;    /* size of allocated array */
-    void  *data;
-};
-
-static const struct array empty_array;
-
-static inline void *array_grow( struct array *array, size_t n, size_t elem_size )
-{
-    if (array->count + n > array->size)
-    {
-        array->size = max( array->size * 2, 16 );
-        array->size = max( array->size, array->count + n );
-        array->data = xrealloc( array->data, array->size * elem_size );
-    }
-    array->count += n;
-    return (char *)array->data + (array->count - n) * elem_size;
-}
-
-#define ARRAY_ADD(array,type) ((type *)array_grow( array, 1, sizeof(type) ))
-#define ARRAY_SORT(array,type,cmp) qsort( (array)->data, (array)->count, sizeof(type), cmp )
-#define ARRAY_ENTRY(array,idx,type) ((type *)(array)->data + (idx))
-#define ARRAY_FOR_EACH(cursor,array,type) \
-    for (type *cursor = (array)->data; cursor - (type *)(array)->data < (array)->count; cursor++)
-
 /* string array functions */
 
 struct strarray
@@ -225,10 +175,6 @@ struct strarray
 
 static const struct strarray empty_strarray;
 
-#define STRARRAY_FOR_EACH(cursor,array) \
-    for (const char **__p = (array)->str, *cursor = NULL; \
-         __p - (array)->str < (array)->count && (cursor = *__p, 1); __p++)
-
 static inline void strarray_add( struct strarray *array, const char *str )
 {
     if (array->count == array->size)
@@ -242,23 +188,29 @@ static inline void strarray_add( struct strarray *array, const char *str )
 
 static inline void strarray_addall( struct strarray *array, struct strarray added )
 {
-    STRARRAY_FOR_EACH( ptr, &added ) strarray_add( array, ptr );
+    unsigned int i;
+
+    for (i = 0; i < added.count; i++) strarray_add( array, added.str[i] );
 }
 
-static inline bool strarray_exists( struct strarray array, const char *str )
+static inline int strarray_exists( const struct strarray *array, const char *str )
 {
-    STRARRAY_FOR_EACH( ptr, &array ) if (!strcmp( ptr, str )) return true;
-    return false;
+    unsigned int i;
+
+    for (i = 0; i < array->count; i++) if (!strcmp( array->str[i], str )) return 1;
+    return 0;
 }
 
 static inline void strarray_add_uniq( struct strarray *array, const char *str )
 {
-    if (!strarray_exists( *array, str )) strarray_add( array, str );
+    if (!strarray_exists( array, str )) strarray_add( array, str );
 }
 
 static inline void strarray_addall_uniq( struct strarray *array, struct strarray added )
 {
-    STRARRAY_FOR_EACH( ptr, &added ) strarray_add_uniq( array, ptr );
+    unsigned int i;
+
+    for (i = 0; i < added.count; i++) strarray_add_uniq( array, added.str[i] );
 }
 
 static inline struct strarray strarray_fromstring( const char *str, const char *delim )
@@ -289,7 +241,7 @@ static inline char *strarray_tostring( struct strarray array, const char *sep )
     unsigned int i, len = 1 + (array.count - 1) * strlen(sep);
 
     if (!array.count) return xstrdup("");
-    STRARRAY_FOR_EACH( str, &array ) len += strlen( str );
+    for (i = 0; i < array.count; i++) len += strlen( array.str[i] );
     str = xmalloc( len );
     strcpy( str, array.str[0] );
     for (i = 1; i < array.count; i++)
@@ -305,12 +257,12 @@ static inline void strarray_qsort( struct strarray *array, int (*func)(const cha
     if (array->count) qsort( array->str, array->count, sizeof(*array->str), (void *)func );
 }
 
-static inline const char *strarray_bsearch( struct strarray array, const char *str,
+static inline const char *strarray_bsearch( const struct strarray *array, const char *str,
                                             int (*func)(const char **, const char **) )
 {
     char **res = NULL;
 
-    if (array.count) res = bsearch( &str, array.str, array.count, sizeof(*array.str), (void *)func );
+    if (array->count) res = bsearch( &str, array->str, array->count, sizeof(*array->str), (void *)func );
     return res ? *res : NULL;
 }
 
@@ -322,9 +274,8 @@ static inline void strarray_trace( struct strarray args )
     {
         if (strpbrk( args.str[i], " \t\n\r")) printf( "\"%s\"", args.str[i] );
         else printf( "%s", args.str[i] );
-        if (i < args.count - 1) putchar( ' ' );
+        putchar( i < args.count - 1 ? ' ' : '\n' );
     }
-    putchar( '\n' );
 }
 
 static inline int strarray_spawn( struct strarray args )
@@ -419,11 +370,12 @@ static inline char *build_relative_path( const char *base, const char *from, con
 extern const char *temp_dir;
 extern struct strarray temp_files;
 
-static inline char *make_temp_dir( const char *tmpdir )
+static inline char *make_temp_dir(void)
 {
     unsigned int value = time(NULL) + getpid();
     int count;
     char *name;
+    const char *tmpdir = NULL;
 
     for (count = 0; count < 0x8000; count++)
     {
@@ -437,7 +389,6 @@ static inline char *make_temp_dir( const char *tmpdir )
         {
             if (!(tmpdir = getenv("TMPDIR"))) tmpdir = "/tmp";
         }
-        else if (errno != EEXIST) fatal_perror( "cannot create directory in %s", tmpdir ? tmpdir : "." );
         free( name );
     }
     fprintf( stderr, "failed to create directory for temp files\n" );
@@ -450,7 +401,7 @@ static inline char *make_temp_file( const char *prefix, const char *suffix )
     int fd, count;
     char *name;
 
-    if (!temp_dir) temp_dir = make_temp_dir( NULL );
+    if (!temp_dir) temp_dir = make_temp_dir();
     if (!suffix) suffix = "";
     if (!prefix) prefix = "tmp";
     else prefix = get_basename_noext( prefix );
@@ -485,7 +436,9 @@ static inline char *make_temp_file( const char *prefix, const char *suffix )
 
 static inline void remove_temp_files(void)
 {
-    STRARRAY_FOR_EACH( file, &temp_files ) if (file) unlink( file );
+    unsigned int i;
+
+    for (i = 0; i < temp_files.count; i++) if (temp_files.str[i]) unlink( temp_files.str[i] );
     if (temp_dir) rmdir( temp_dir );
 }
 
@@ -500,29 +453,6 @@ static inline void init_signals( void (*cleanup)(int) )
 }
 
 
-static inline void mkdir_p( const char *dir )
-{
-    char *p, *path;
-
-    if (!dir[0]) return;
-
-    path = xstrdup( dir );
-    for (p = path + 1; (p = strchr( p, '/' )); p++)
-    {
-        *p = 0;
-        if (mkdir( path, 0755 ) == -1 && errno != EEXIST) fatal_perror( "mkdir %s", path );
-        *p = '/';
-    }
-    if (mkdir( path, 0755 ) == -1)
-    {
-        struct stat st;
-        if (errno != EEXIST || stat( path, &st ) || !S_ISDIR( st.st_mode ))
-            fatal_perror( "cannot create %s", path );
-    }
-    free( path );
-}
-
-
 static inline void *read_file( const char *name, size_t *size )
 {
     struct stat st;
@@ -617,14 +547,6 @@ static inline void set_target_ptr_size( struct target *target, unsigned int size
 }
 
 
-static inline bool is_pe_target( struct target target )
-{
-    return (target.platform == PLATFORM_WINDOWS ||
-            target.platform == PLATFORM_MINGW ||
-            target.platform == PLATFORM_CYGWIN);
-}
-
-
 static inline int get_cpu_from_name( const char *name )
 {
     static const struct
@@ -694,10 +616,19 @@ static inline const char *get_arch_dir( struct target target )
     };
 
     if (!cpu_names[target.cpu]) return "";
-    return strmake( "/%s-%s", cpu_names[target.cpu], is_pe_target( target ) ? "windows" : "unix" );
+
+    switch (target.platform)
+    {
+    case PLATFORM_WINDOWS:
+    case PLATFORM_CYGWIN:
+    case PLATFORM_MINGW:
+        return strmake( "/%s-windows", cpu_names[target.cpu] );
+    default:
+        return strmake( "/%s-unix", cpu_names[target.cpu] );
+    }
 }
 
-static inline bool parse_target( const char *name, struct target *target )
+static inline int parse_target( const char *name, struct target *target )
 {
     int res;
     char *p, *spec = xstrdup( name );
@@ -712,7 +643,7 @@ static inline bool parse_target( const char *name, struct target *target )
         if ((res = get_cpu_from_name( spec )) == -1)
         {
             free( spec );
-            return false;
+            return 0;
         }
         target->cpu = res;
     }
@@ -724,7 +655,7 @@ static inline bool parse_target( const char *name, struct target *target )
     else
     {
         free( spec );
-        return false;
+        return 0;
     }
 
     /* get the OS part */
@@ -742,7 +673,7 @@ static inline bool parse_target( const char *name, struct target *target )
     }
 
     free( spec );
-    return true;
+    return 1;
 }
 
 
@@ -774,12 +705,6 @@ static inline char *get_bindir( const char *argv0 )
     if (!sysctl( pathname, ARRAY_SIZE(pathname), path, &path_size, NULL, 0 ))
         dir = realpath( path, NULL );
     free( path );
-#elif defined(__APPLE__)
-    uint32_t path_size = PATH_MAX;
-    char *path = xmalloc( path_size );
-    if (!_NSGetExecutablePath( path, &path_size ))
-        dir = realpath( path, NULL );
-    free( path );
 #endif
     if (!dir && !(dir = realpath( argv0, NULL ))) return NULL;
     return get_dirname( dir );
@@ -921,7 +846,7 @@ static inline void flush_output_buffer( const char *name )
 struct long_option
 {
     const char *name;
-    bool has_arg;
+    int has_arg;
     int val;
 };
 
