diff --git a/tools/widl/client.c b/tools/widl/client.c
index e5d93cdc1c395361a62024ccfa41e447d7678f2d..795e3ba2104d1193fe770b3245df0a1f853a9baa 100644
--- a/tools/widl/client.c
+++ b/tools/widl/client.c
@@ -439,11 +439,9 @@ static void write_stubdescriptor(type_t *iface, int expr_eval_routines)
 
 static void write_clientinterfacedecl(type_t *iface)
 {
+    unsigned int ver = get_attrv(iface->attrs, ATTR_VERSION);
     const struct uuid *uuid = get_attrp(iface->attrs, ATTR_UUID);
     const str_list_t *endpoints = get_attrp(iface->attrs, ATTR_ENDPOINT);
-    unsigned short major, minor;
-
-    get_version( iface->attrs, &major, &minor );
 
     if (endpoints) write_endpoints( client, iface->name, endpoints );
 
@@ -454,7 +452,7 @@ static void write_clientinterfacedecl(type_t *iface)
     print_client("{{0x%08x,0x%04x,0x%04x,{0x%02x,0x%02x,0x%02x,0x%02x,0x%02x,0x%02x,0x%02x,0x%02x}},{%d,%d}},\n",
                  uuid->Data1, uuid->Data2, uuid->Data3, uuid->Data4[0], uuid->Data4[1],
                  uuid->Data4[2], uuid->Data4[3], uuid->Data4[4], uuid->Data4[5], uuid->Data4[6],
-                 uuid->Data4[7], major, minor);
+                 uuid->Data4[7], MAJORVERSION(ver), MINORVERSION(ver));
     print_client("{{0x8a885d04,0x1ceb,0x11c9,{0x9f,0xe8,0x08,0x00,0x2b,0x10,0x48,0x60}},{2,0}},\n"); /* FIXME */
     print_client("0,\n");
     if (endpoints)
@@ -477,7 +475,7 @@ static void write_clientinterfacedecl(type_t *iface)
                      iface->name, iface->name);
     else
         print_client("RPC_IF_HANDLE %s%s_v%d_%d_c_ifspec = (RPC_IF_HANDLE)& %s___RpcClientInterface;\n",
-                     prefix_client, iface->name, major, minor, iface->name);
+                     prefix_client, iface->name, MAJORVERSION(ver), MINORVERSION(ver), iface->name);
     fprintf(client, "\n");
 }
 
