diff --git a/tools/widl/server.c b/tools/widl/server.c
index 4c72b106069f0dd43276a5bb3f7871512a8e5f06..8bc179f14388fe7e24aa9220c027ad2587030684 100644
--- a/tools/widl/server.c
+++ b/tools/widl/server.c
@@ -266,11 +266,9 @@ static void write_function_stubs(type_t *iface, unsigned int *proc_offset)
 
 static void write_dispatchtable(type_t *iface)
 {
+    unsigned int ver = get_attrv(iface->attrs, ATTR_VERSION);
     unsigned int method_count = 0;
     const statement_t *stmt;
-    unsigned short major, minor;
-
-    get_version( iface->attrs, &major, &minor );
 
     print_server("static RPC_DISPATCH_FUNCTION %s_table[] =\n", iface->name);
     print_server("{\n");
@@ -288,7 +286,7 @@ static void write_dispatchtable(type_t *iface)
     print_server("0\n");
     indent--;
     print_server("};\n");
-    print_server("static RPC_DISPATCH_TABLE %s_v%d_%d_DispatchTable =\n", iface->name, major, minor);
+    print_server("static RPC_DISPATCH_TABLE %s_v%d_%d_DispatchTable =\n", iface->name, MAJORVERSION(ver), MINORVERSION(ver));
     print_server("{\n");
     indent++;
     print_server("%u,\n", method_count);
@@ -405,15 +403,13 @@ static void write_stubdescriptor(type_t *iface, int expr_eval_routines)
 
 static void write_serverinterfacedecl(type_t *iface)
 {
+    unsigned int ver = get_attrv(iface->attrs, ATTR_VERSION);
     struct uuid *uuid = get_attrp(iface->attrs, ATTR_UUID);
     const str_list_t *endpoints = get_attrp(iface->attrs, ATTR_ENDPOINT);
-    unsigned short major, minor;
-
-    get_version( iface->attrs, &major, &minor );
 
     if (endpoints) write_endpoints( server, iface->name, endpoints );
 
-    print_server("static RPC_DISPATCH_TABLE %s_v%d_%d_DispatchTable;\n", iface->name, major, minor);
+    print_server("static RPC_DISPATCH_TABLE %s_v%d_%d_DispatchTable;\n", iface->name, MAJORVERSION(ver), MINORVERSION(ver));
     print_server( "static const MIDL_SERVER_INFO %s_ServerInfo;\n", iface->name );
     fprintf(server, "\n");
     print_server("static const RPC_SERVER_INTERFACE %s___RpcServerInterface =\n", iface->name );
@@ -423,9 +419,9 @@ static void write_serverinterfacedecl(type_t *iface)
     print_server("{{0x%08x,0x%04x,0x%04x,{0x%02x,0x%02x,0x%02x,0x%02x,0x%02x,0x%02x,0x%02x,0x%02x}},{%d,%d}},\n",
                  uuid->Data1, uuid->Data2, uuid->Data3, uuid->Data4[0], uuid->Data4[1],
                  uuid->Data4[2], uuid->Data4[3], uuid->Data4[4], uuid->Data4[5], uuid->Data4[6],
-                 uuid->Data4[7], major, minor);
+                 uuid->Data4[7], MAJORVERSION(ver), MINORVERSION(ver));
     print_server("{{0x8a885d04,0x1ceb,0x11c9,{0x9f,0xe8,0x08,0x00,0x2b,0x10,0x48,0x60}},{2,0}},\n"); /* FIXME */
-    print_server("&%s_v%d_%d_DispatchTable,\n", iface->name, major, minor);
+    print_server("&%s_v%d_%d_DispatchTable,\n", iface->name, MAJORVERSION(ver), MINORVERSION(ver));
     if (endpoints)
     {
         print_server("%u,\n", list_count(endpoints));
@@ -446,7 +442,7 @@ static void write_serverinterfacedecl(type_t *iface)
                      iface->name, iface->name);
     else
         print_server("RPC_IF_HANDLE %s%s_v%d_%d_s_ifspec = (RPC_IF_HANDLE)& %s___RpcServerInterface;\n",
-                     prefix_server, iface->name, major, minor, iface->name);
+                     prefix_server, iface->name, MAJORVERSION(ver), MINORVERSION(ver), iface->name);
     fprintf(server, "\n");
 }
 
