diff --git a/tools/widl/typetree.h b/tools/widl/typetree.h
index d49d7fe524405b7a95fbcf28858bc7078a2f3be5..6e5c606ab351a22404e3b4a379d8956b88d808d4 100644
--- a/tools/widl/typetree.h
+++ b/tools/widl/typetree.h
@@ -84,26 +84,9 @@ type_t *type_parameterized_type_specialize_declare(type_t *type, typeref_list_t
 type_t *type_parameterized_type_specialize_define(type_t *type);
 int type_is_equal(const type_t *type1, const type_t *type2);
 const char *type_get_decl_name(const type_t *type, enum name_type name_type);
-extern const char *type_get_record_specifier( type_t *type );
-extern const char *type_get_name( const type_t *type, enum name_type name_type, bool record );
+const char *type_get_name(const type_t *type, enum name_type name_type);
 char *gen_name(void);
 
-extern bool needs_space_after( const type_t *type );
-extern bool decl_needs_parens( const type_t *type );
-
-extern void append_basic_type( struct strbuf *str, const type_t *type );
-extern void append_pointer_left( struct strbuf *str, const type_t *type, const char *callconv );
-extern void append_type_left( struct strbuf *str, const decl_spec_t *decl_spec, enum name_type name_type,
-                              const char *callconv );
-extern void append_type_right( struct strbuf *str, const type_t *type, const char *callconv, bool is_field );
-extern void append_declspec( struct strbuf *str, const decl_spec_t *decl_spec, enum name_type name_type,
-                             const char *callconv, bool is_field, const char *name );
-
-extern char *format_namespace( const struct namespace *namespace, const char *prefix, const char *separator,
-                               const char *suffix, const char *abi_prefix );
-extern char *format_parameterized_type_name( const type_t *type, const typeref_list_t *params );
-extern char *format_parameterized_type_args( const type_t *type, const char *prefix, const char *suffix );
-
 typeref_t *make_typeref(type_t *type);
 typeref_list_t *append_typeref(typeref_list_t *list, typeref_t *ref);
 
