diff --git a/tools/widl/utils.c b/tools/widl/utils.c
index 8d2758d9bea1212497960c470475a03c2cd67b35..4f2e9d3d602a6fa8efc5242e607d02f32ea7dbf8 100644
--- a/tools/widl/utils.c
+++ b/tools/widl/utils.c
@@ -117,25 +117,41 @@ size_t widl_getline(char **linep, size_t *lenp, FILE *fp)
     return n;
 }
 
-void strappend( struct strbuf *str, const char *fmt, ... )
+size_t strappend(char **buf, size_t *len, size_t pos, const char* fmt, ...)
 {
+    size_t size;
     va_list ap;
+    char *ptr;
     int n;
 
-    assert( (str->len == 0 && str->buf == NULL) ||
-            (str->len != 0 && str->buf != NULL) );
+    assert( buf && len );
+    assert( (*len == 0 && *buf == NULL) || (*len != 0 && *buf != NULL) );
+
+    if (*buf)
+    {
+        size = *len;
+        ptr = *buf;
+    }
+    else
+    {
+        size = 100;
+        ptr = xmalloc( size );
+    }
 
     for (;;)
     {
         va_start( ap, fmt );
-        n = str->len ? vsnprintf( str->buf + str->pos, str->len - str->pos, fmt, ap ) : 128;
+        n = vsnprintf( ptr + pos, size - pos, fmt, ap );
         va_end( ap );
-        if (n >= 0 && n <= str->len && str->pos + n < str->len) break;
-        str->len = max( str->pos + n, str->len * 3 / 2 );
-        str->buf = xrealloc( str->buf, str->len );
+        if (n == -1) size *= 2;
+        else if (pos + (size_t)n >= size) size = pos + n + 1;
+        else break;
+        ptr = xrealloc( ptr, size );
     }
 
-    str->pos += n;
+    *len = size;
+    *buf = ptr;
+    return n;
 }
 
 /*******************************************************************
