diff --git a/tools/widl/utils.h b/tools/widl/utils.h
index 538121629a1fc4f33f9f09c3f158ba7266127f08..2a6bc7d793054b7dfe1351ae8e5407b4a5101fda 100644
--- a/tools/widl/utils.h
+++ b/tools/widl/utils.h
@@ -31,14 +31,7 @@ void warning(const char *s, ...) __attribute__((format (printf, 1, 2)));
 void warning_at( const struct location *, const char *s, ... ) __attribute__((format( printf, 2, 3 )));
 #define warning_loc( ... ) warning_at( NULL, ## __VA_ARGS__ )
 void chat(const char *s, ...) __attribute__((format (printf, 1, 2)));
-
-struct strbuf
-{
-    char *buf;
-    size_t pos;
-    size_t len;
-};
-extern void strappend( struct strbuf *str, const char *fmt, ... ) __attribute__((__format__( __printf__, 2, 3 )));
+size_t strappend(char **buf, size_t *len, size_t pos, const char* fmt, ...) __attribute__((__format__ (__printf__, 4, 5 )));
 
 size_t widl_getline(char **linep, size_t *lenp, FILE *fp);
 
@@ -48,4 +41,11 @@ extern void add_output_to_resources( const char *type, const char *name );
 extern void flush_output_resources( const char *name );
 extern void put_pword( unsigned int val );
 extern void put_str( int indent, const char *format, ... ) __attribute__((format (printf, 2, 3)));
+
+/* typelibs expect the minor version to be stored in the higher bits and
+ * major to be stored in the lower bits */
+#define MAKEVERSION(major, minor) ((((minor) & 0xffff) << 16) | ((major) & 0xffff))
+#define MAJORVERSION(version) ((version) & 0xffff)
+#define MINORVERSION(version) (((version) >> 16) & 0xffff)
+
 #endif
