diff --git a/tools/widl/widl.man.in b/tools/widl/widl.man.in
index 6e483c33927b49e6d6fa6c1ed6457dbbf751b71e..7e234de70bb6f2765136d1089f926138b668ce58 100644
--- a/tools/widl/widl.man.in
+++ b/tools/widl/widl.man.in
@@ -107,12 +107,6 @@ binary resource file containing the script is generated instead.
 Regenerate the dlldata file from scratch using the specified proxy
 names. The default output filename is \fBdlldata.c\fR.
 .PP
-.B Metadata file options:
-.IP "\fB--winmd\fR"
-Generate a Windows Runtime Metadata file. This implies the
-\fB--winrt\fR mode. The default output filename is
-\fIinfile\fB.winmd\fR.
-.PP
 .B Preprocessor options:
 .IP "\fB-I \fIpath\fR"
 Add a directory to the include search path. Multiple include
