diff --git a/tools/widl/widltypes.h b/tools/widl/widltypes.h
index 03b4daf2aa9a95ec6f7cbf919da066fd2fa45d6c..5256e003c6f360066c49ec74b4cd97e9730f91cc 100644
--- a/tools/widl/widltypes.h
+++ b/tools/widl/widltypes.h
@@ -55,7 +55,6 @@ typedef struct _user_type_t context_handle_t;
 typedef struct _user_type_t generic_handle_t;
 typedef struct _statement_t statement_t;
 typedef struct _warning_t warning_t;
-typedef struct _version_t version_t;
 
 typedef struct list attr_list_t;
 typedef struct list str_list_t;
@@ -75,7 +74,6 @@ enum attr_type
     ATTR_AGGREGATABLE,
     ATTR_ALLOCATE,
     ATTR_ANNOTATION,
-    ATTR_APICONTRACT, /* implicit attribute */
     ATTR_APPOBJECT,
     ATTR_ASYNC,
     ATTR_ASYNCUUID,
@@ -345,8 +343,6 @@ struct _attr_t {
     unsigned int ival;
     void *pval;
   } u;
-  /* metadata */
-  unsigned int md_member;
   /* parser-internal */
   struct list entry;
   struct location where;
@@ -368,7 +364,7 @@ struct _expr_t {
     double dval;
     const char *sval;
     const expr_t *ext;
-    var_t *var;
+    decl_spec_t tref;
   } u;
   const expr_t *ext2;
   int is_const;
@@ -406,7 +402,6 @@ struct iface_details
   struct _type_t *inherit;
   struct _type_t *disp_inherit;
   struct _type_t *async_iface;
-  struct _type_t *runtime_class;
   typeref_list_t *requires;
 };
 
@@ -501,23 +496,6 @@ enum type_type
     TYPE_DELEGATE,
 };
 
-struct metadata
-{
-    unsigned int name;
-    unsigned int namespace;
-    unsigned int ref;
-    unsigned int def;
-    unsigned int extends;
-    /* get/put methods */
-    unsigned int class_property;
-    unsigned int iface_property;
-    unsigned int propertymap;
-    /* add/remove methods */
-    unsigned int class_event;
-    unsigned int iface_event;
-    unsigned int eventmap;
-};
-
 struct _type_t {
   const char *name;               /* C++ name with parameters in brackets */
   struct namespace *namespace;
@@ -546,11 +524,9 @@ struct _type_t {
   const char *impl_name;          /* C++ parameterized types impl base class name */
   const char *param_name;         /* used to build c_name of a parameterized type, when used as a parameter */
   const char *short_name;         /* widl specific short name */
-  const char *winmd_short_name;   /* metadata short name for parameterized type */
   unsigned int typestring_offset;
   unsigned int ptrdesc;           /* used for complex structs */
   int typelib_idx;
-  struct metadata md;
   struct location where;
   unsigned int ignore : 1;
   unsigned int defined : 1;
@@ -662,11 +638,6 @@ struct _statement_t {
     unsigned int is_defined : 1;
 };
 
-struct _version_t {
-    unsigned short major;
-    unsigned short minor;
-};
-
 struct _warning_t {
     int num;
     struct list entry;
@@ -702,6 +673,10 @@ type_t *reg_type(type_t *type, const char *name, struct namespace *namespace, in
 var_t *make_var(char *name);
 var_list_t *append_var(var_list_t *list, var_t *var);
 
+char *format_namespace(struct namespace *namespace, const char *prefix, const char *separator, const char *suffix,
+                       const char *abi_prefix);
+char *format_parameterized_type_name(type_t *type, typeref_list_t *params);
+
 static inline enum type_type type_get_type_detect_alias(const type_t *type)
 {
     return type->type_type;
@@ -712,11 +687,6 @@ static inline enum type_type type_get_type_detect_alias(const type_t *type)
     if (stmt->type == STMT_DECLARATION && stmt->u.var->declspec.stgclass == STG_NONE && \
         type_get_type_detect_alias(stmt->u.var->declspec.type) == TYPE_FUNCTION)
 
-#define STATEMENTS_FOR_EACH_FUNC_REV(stmt, stmts) \
-  if (stmts) LIST_FOR_EACH_ENTRY_REV( stmt, stmts, statement_t, entry ) \
-    if (stmt->type == STMT_DECLARATION && stmt->u.var->declspec.stgclass == STG_NONE && \
-        type_get_type_detect_alias(stmt->u.var->declspec.type) == TYPE_FUNCTION)
-
 static inline int statements_has_func(const statement_list_t *stmts)
 {
   const statement_t *stmt;
