diff --git a/tools/widl/write_sltg.c b/tools/widl/write_sltg.c
index 3c80e21d86bd06361c84473d1c17f466708066c3..414a8ec6855347816eaf08a467ed29af00926ad3 100644
--- a/tools/widl/write_sltg.c
+++ b/tools/widl/write_sltg.c
@@ -85,7 +85,7 @@ struct sltg_hrefmap
     int *href;
 };
 
-#pragma pack(push,1)
+#include "pshpack1.h"
 struct sltg_typeinfo_header
 {
     short magic;
@@ -221,7 +221,7 @@ struct sltg_impl_info
     short pos;
 };
 
-#pragma pack(pop)
+#include "poppack.h"
 
 static void add_structure_typeinfo(struct sltg_typelib *typelib, type_t *type);
 static void add_interface_typeinfo(struct sltg_typelib *typelib, type_t *type);
@@ -320,13 +320,8 @@ static void init_library(struct sltg_typelib *sltg)
         switch (attr->type)
         {
         case ATTR_VERSION:
-        {
-            const version_t *version = attr->u.pval;
-            unsigned short major = version ? version->major : 0, minor = version ? version->minor : 0;
-
-            sltg->library.version = (minor << 16) | major;
+            sltg->library.version = attr->u.ival;
             break;
-        }
         case ATTR_HELPSTRING:
             sltg->library.helpstring = attr->u.pval;
             break;
@@ -527,16 +522,12 @@ static const char *add_typeinfo_block(struct sltg_typelib *typelib, const type_t
 static void init_typeinfo(struct sltg_typeinfo_header *ti, const type_t *type, short kind,
                           const struct sltg_hrefmap *hrefmap)
 {
-    unsigned short major, minor;
-
-    get_version( type->attrs, &major, &minor );
-
     ti->magic = 0x0501;
     ti->href_offset = -1;
     ti->res06 = -1;
     ti->member_offset = sizeof(*ti);
     ti->res0e = -1;
-    ti->version = minor << 16 | major;
+    ti->version = get_attrv(type->attrs, ATTR_VERSION);
     ti->res16 = 0xfffe0000;
     ti->misc.unknown1 = 0x02;
     ti->misc.flags = 0; /* FIXME */
