diff --git a/tools/wine/Makefile.in b/tools/wine/Makefile.in
deleted file mode 100644
index 55489a3444afadc0a23fe29c1dfc51347dd838cc..0000000000000000000000000000000000000000
--- a/tools/wine/Makefile.in
+++ /dev/null
@@ -1,12 +0,0 @@
-PROGRAMS = wine
-
-SOURCES = \
-	wine.c \
-	wine.de.UTF-8.man.in \
-	wine.fr.UTF-8.man.in \
-	wine.man.in \
-	wine.pl.UTF-8.man.in
-
-wine_EXTRADEFS = -DBINDIR="\"${bindir}\"" -DLIBDIR="\"${libdir}\""
-
-INSTALL_LIB = $(PROGRAMS)
