diff --git a/tools/winebuild/build.h b/tools/winebuild/build.h
index 3acf987aab29986ca0610afb89218b2805c2aab3..e5ede93425baafef6813bf692955d20e4c932b6f 100644
--- a/tools/winebuild/build.h
+++ b/tools/winebuild/build.h
@@ -65,6 +65,12 @@ enum arg_type
 
 #define MAX_ARGUMENTS 32
 
+typedef struct
+{
+    int n_values;
+    unsigned int *values;
+} ORD_VARIABLE;
+
 typedef struct
 {
     int           nb_args;
@@ -72,6 +78,11 @@ typedef struct
     enum arg_type args[MAX_ARGUMENTS];
 } ORD_FUNCTION;
 
+typedef struct
+{
+    unsigned short value;
+} ORD_ABS;
+
 typedef struct
 {
     ORD_TYPE    type;
@@ -84,9 +95,9 @@ typedef struct
     char       *export_name;  /* name exported under for noname exports */
     union
     {
-        struct array   var;
-        unsigned short abs;
+        ORD_VARIABLE   var;
         ORD_FUNCTION   func;
+        ORD_ABS        abs;
     } u;
 } ORDDEF;
 
@@ -108,7 +119,9 @@ struct apiset_entry
 
 struct apiset
 {
-    struct array entries;
+    unsigned int count;
+    unsigned int size;
+    struct apiset_entry *entries;
     unsigned int str_pos;
     unsigned int str_size;
     char *strings;
@@ -138,16 +151,19 @@ typedef struct
     SPEC_TYPE        type;               /* type of dll (Win16/Win32) */
     int              stack_size;         /* exe stack size */
     int              heap_size;          /* exe heap size */
+    int              nb_entry_points;    /* number of used entry points */
+    int              alloc_entry_points; /* number of allocated entry points */
+    unsigned int     nb_resources;       /* number of resources */
     int              characteristics;    /* characteristics for the PE header */
     int              dll_characteristics;/* DLL characteristics for the PE header */
     int              subsystem;          /* subsystem id */
     int              subsystem_major;    /* subsystem version major number */
     int              subsystem_minor;    /* subsystem version minor number */
     int              unicode_app;        /* default to unicode entry point */
-    struct array     entry_points;       /* spec entry points */
+    ORDDEF          *entry_points;       /* spec entry points */
     struct exports   exports;            /* dll exports */
     struct exports   native_exports;     /* dll native exports */
-    struct array     resources;          /* array of dll resources (format differs between Win16/Win32) */
+    struct resource *resources;          /* array of dll resources (format differs between Win16/Win32) */
     struct apiset    apiset;             /* list of defined api sets */
 } DLLSPEC;
 
@@ -161,7 +177,7 @@ static inline unsigned int get_ptr_size(void)
 
 static inline int is_pe(void)
 {
-    return is_pe_target( target );
+    return target.platform == PLATFORM_MINGW || target.platform == PLATFORM_WINDOWS;
 }
 
 /* entry point flags */
@@ -237,6 +253,8 @@ static inline int is_pe(void)
 extern char *strupper(char *s);
 extern DECLSPEC_NORETURN void fatal_error( const char *msg, ... )
    __attribute__ ((__format__ (__printf__, 1, 2)));
+extern DECLSPEC_NORETURN void fatal_perror( const char *msg, ... )
+   __attribute__ ((__format__ (__printf__, 1, 2)));
 extern void error( const char *msg, ... )
    __attribute__ ((__format__ (__printf__, 1, 2)));
 extern void warning( const char *msg, ... )
@@ -267,11 +285,13 @@ extern void dump_bytes( const void *buffer, unsigned int size );
 extern int remove_stdcall_decoration( char *name );
 extern void assemble_file( const char *src_file, const char *obj_file );
 extern DLLSPEC *alloc_dll_spec(void);
+extern void free_dll_spec( DLLSPEC *spec );
 extern char *make_c_identifier( const char *str );
+extern const char *get_stub_name( const ORDDEF *odp, const DLLSPEC *spec );
 extern const char *get_abi_name( const ORDDEF *odp, const char *name );
 extern const char *get_link_name( const ORDDEF *odp );
 extern int sort_func_list( ORDDEF **list, int count, int (*compare)(const void *, const void *) );
-extern unsigned int get_section_alignment(void);
+extern unsigned int get_page_size(void);
 extern unsigned int get_args_size( const ORDDEF *odp );
 extern const char *asm_name( const char *func );
 extern const char *arm64_name( const char *func );
@@ -367,7 +387,6 @@ extern struct strarray as_command;
 extern struct strarray cc_command;
 extern struct strarray ld_command;
 extern struct strarray nm_command;
-extern struct strarray strip_command;
 extern char *cpu_option;
 extern char *fpu_option;
 extern char *arch_option;
