diff --git a/tools/winebuild/parser.c b/tools/winebuild/parser.c
index b2a85520766b248edcdb4a453b2fbc9f587cc7f6..ec42c240a0a7039bc292751304fe7ca149faa979 100644
--- a/tools/winebuild/parser.c
+++ b/tools/winebuild/parser.c
@@ -163,8 +163,15 @@ static const char * GetToken( int allow_eol )
 
 static ORDDEF *add_entry_point( DLLSPEC *spec )
 {
-    ORDDEF *ret = ARRAY_ADD( &spec->entry_points, ORDDEF );
+    ORDDEF *ret;
 
+    if (spec->nb_entry_points == spec->alloc_entry_points)
+    {
+        spec->alloc_entry_points += 128;
+        spec->entry_points = xrealloc( spec->entry_points,
+                                       spec->alloc_entry_points * sizeof(*spec->entry_points) );
+    }
+    ret = &spec->entry_points[spec->nb_entry_points++];
     memset( ret, 0, sizeof(*ret) );
     return ret;
 }
@@ -177,8 +184,10 @@ static ORDDEF *add_entry_point( DLLSPEC *spec )
 static int parse_spec_variable( ORDDEF *odp, DLLSPEC *spec )
 {
     char *endptr;
+    unsigned int *value_array;
+    int n_values;
+    int value_array_size;
     const char *token;
-    struct array values = empty_array;
 
     if (spec->type == SPEC_WIN32)
     {
@@ -193,20 +202,38 @@ static int parse_spec_variable( ORDDEF *odp, DLLSPEC *spec )
         return 0;
     }
 
+    n_values = 0;
+    value_array_size = 25;
+    value_array = xmalloc(sizeof(*value_array) * value_array_size);
+
     for (;;)
     {
-        if (!(token = GetToken(0))) return 0;
-	if (*token == ')') break;
+        if (!(token = GetToken(0)))
+        {
+            free( value_array );
+            return 0;
+        }
+	if (*token == ')')
+	    break;
+
+	value_array[n_values++] = strtoul(token, &endptr, 0);
+	if (n_values == value_array_size)
+	{
+	    value_array_size += 25;
+	    value_array = xrealloc(value_array,
+				   sizeof(*value_array) * value_array_size);
+	}
 
-        *ARRAY_ADD( &values, unsigned int ) = strtoul(token, &endptr, 0);
 	if (endptr == NULL || *endptr != '\0')
         {
             error( "Expected number value, got '%s'\n", token );
+            free( value_array );
             return 0;
         }
     }
 
-    odp->u.var = values;
+    odp->u.var.n_values = n_values;
+    odp->u.var.values = xrealloc(value_array, sizeof(*value_array) * n_values);
     return 1;
 }
 
@@ -222,8 +249,6 @@ static int parse_spec_arguments( ORDDEF *odp, DLLSPEC *spec, int optional )
     unsigned int i, arg;
     int is_win32 = (spec->type == SPEC_WIN32) || (odp->flags & FLAG_EXPORT32);
 
-    odp->u.func.nb_args = 0;
-
     if (!(token = GetToken( optional ))) return optional;
     if (*token != '(')
     {
@@ -231,6 +256,7 @@ static int parse_spec_arguments( ORDDEF *odp, DLLSPEC *spec, int optional )
         return 0;
     }
 
+    odp->u.func.nb_args = 0;
     for (i = 0; i < MAX_ARGUMENTS; i++)
     {
         if (!(token = GetToken(0))) return 0;
@@ -281,11 +307,6 @@ static int parse_spec_arguments( ORDDEF *odp, DLLSPEC *spec, int optional )
             error( "First argument of a thiscall function must be a pointer\n" );
             return 0;
         }
-        if (odp->flags & FLAG_CPU_MASK & ~FLAG_CPU(CPU_i386))
-        {
-            error( "A thiscall function can only be exported on i386\n" );
-            return 0;
-        }
     }
     if (odp->flags & FLAG_FASTCALL)
     {
@@ -300,7 +321,7 @@ static int parse_spec_arguments( ORDDEF *odp, DLLSPEC *spec, int optional )
     }
     if (odp->flags & FLAG_SYSCALL)
     {
-        if (odp->type != TYPE_STDCALL && odp->type != TYPE_STUB)
+        if (odp->type != TYPE_STDCALL)
         {
             error( "A syscall function must use the stdcall convention\n" );
             return 0;
@@ -337,7 +358,7 @@ static int parse_spec_export( ORDDEF *odp, DLLSPEC *spec )
         odp->flags |= FLAG_NORELAY;  /* no relay debug possible for varags entry point */
 
     if (target.cpu != CPU_i386)
-        odp->flags &= ~FLAG_FASTCALL;
+        odp->flags &= ~(FLAG_THISCALL | FLAG_FASTCALL);
 
     if (!(token = GetToken(1)))
     {
@@ -393,7 +414,7 @@ static int parse_spec_equate( ORDDEF *odp, DLLSPEC *spec )
         error( "Value %d for absolute symbol doesn't fit in 16 bits\n", value );
         value = 0;
     }
-    odp->u.abs = value;
+    odp->u.abs.value = value;
     return 1;
 }
 
@@ -405,16 +426,10 @@ static int parse_spec_equate( ORDDEF *odp, DLLSPEC *spec )
  */
 static int parse_spec_stub( ORDDEF *odp, DLLSPEC *spec )
 {
-    const char *token;
-
-    if (!parse_spec_arguments( odp, spec, 1 )) return 0;
+    odp->u.func.nb_args = -1;
+    odp->link_name = xstrdup("");
 
-    if (!(token = GetToken(1)))
-        odp->link_name = xstrdup( odp->name );
-    else
-        odp->link_name = xstrdup( token );
-
-    return 1;
+    return parse_spec_arguments( odp, spec, 1 );
 }
 
 
@@ -492,15 +507,6 @@ static const char *parse_spec_flags( DLLSPEC *spec, ORDDEF *odp, const char *tok
             }
             free( args );
         }
-        else if (!strncmp( token, "syscall=", 8 ))
-        {
-            char *end;
-            unsigned int id = strtoul( token + 8, &end, 0 );
-
-            if (*end || id >= 0x4000)
-                error( "Invalid syscall number '%s', should be in range 0-0x3fff\n", token + 8 );
-            odp->flags |= FLAG_SYSCALL;
-        }
         else if (!strcmp( token, "i386" ))  /* backwards compatibility */
         {
             odp->flags |= FLAG_CPU(CPU_i386);
@@ -684,7 +690,7 @@ static int parse_spec_ordinal( int ordinal, DLLSPEC *spec )
     return 1;
 
 error:
-    spec->entry_points.count--;
+    spec->nb_entry_points--;
     free( odp->name );
     return 0;
 }
@@ -721,20 +727,27 @@ static unsigned int apiset_add_str( struct apiset *apiset, const char *str, unsi
     return ret - apiset->strings;
 }
 
-static struct apiset_entry *add_apiset( struct apiset *apiset, const char *api )
+static void add_apiset( struct apiset *apiset, const char *api )
 {
-    struct apiset_entry *entry = ARRAY_ADD( &apiset->entries, struct apiset_entry );
+    struct apiset_entry *entry;
 
+    if (apiset->count == apiset->size)
+    {
+        apiset->size = max( apiset->size * 2, 64 );
+        apiset->entries = xrealloc( apiset->entries, apiset->size * sizeof(*apiset->entries) );
+    }
+    entry = &apiset->entries[apiset->count++];
     entry->name_len = strlen( api );
     entry->name_off = apiset_add_str( apiset, api, entry->name_len );
     entry->hash = apiset_hash( api );
     entry->hash_len = apiset_hash_len( api );
     entry->val_count = 0;
-    return entry;
 }
 
-static void add_apiset_value( struct apiset *apiset, struct apiset_entry *entry, const char *value )
+static void add_apiset_value( struct apiset *apiset, const char *value )
 {
+    struct apiset_entry *entry = &apiset->entries[apiset->count - 1];
+
     if (entry->val_count < ARRAY_SIZE(entry->values) - 1)
     {
         struct apiset_value *val = &entry->values[entry->val_count++];
@@ -764,7 +777,7 @@ static int parse_spec_apiset( DLLSPEC *spec )
 {
     struct apiset_entry *entry;
     const char *token;
-    unsigned int hash;
+    unsigned int i, hash;
 
     if (!data_only)
     {
@@ -781,7 +794,7 @@ static int parse_spec_apiset( DLLSPEC *spec )
     }
 
     hash = apiset_hash( token );
-    ARRAY_FOR_EACH( entry, &spec->apiset.entries, struct apiset_entry )
+    for (i = 0, entry = spec->apiset.entries; i < spec->apiset.count; i++, entry++)
     {
         if (entry->name_len == strlen( token ) &&
             !strncmp( spec->apiset.strings + entry->name_off, token, entry->name_len ))
@@ -796,8 +809,7 @@ static int parse_spec_apiset( DLLSPEC *spec )
             return 0;
         }
     }
-
-    entry = add_apiset( &spec->apiset, token );
+    add_apiset( &spec->apiset, token );
 
     if (!(token = GetToken(0)) || strcmp( token, "=" ))
     {
@@ -805,7 +817,7 @@ static int parse_spec_apiset( DLLSPEC *spec )
         return 0;
     }
 
-    while ((token = GetToken(1))) add_apiset_value( &spec->apiset, entry, token );
+    while ((token = GetToken(1))) add_apiset_value( &spec->apiset, token );
     return 1;
 }
 
@@ -933,9 +945,12 @@ static void assign_ordinals( struct exports *exports )
 
 static void assign_exports( DLLSPEC *spec, unsigned int cpu, struct exports *exports )
 {
-    exports->entry_points = xmalloc( spec->entry_points.count * sizeof(*exports->entry_points) );
-    ARRAY_FOR_EACH( entry, &spec->entry_points, ORDDEF )
+    unsigned int i;
+
+    exports->entry_points = xmalloc( spec->nb_entry_points * sizeof(*exports->entry_points) );
+    for (i = 0; i < spec->nb_entry_points; i++)
     {
+        ORDDEF *entry = &spec->entry_points[i];
         if ((entry->flags & FLAG_CPU_MASK) && !(entry->flags & FLAG_CPU(cpu)))
             continue;
         exports->entry_points[exports->nb_entry_points++] = entry;
@@ -1223,7 +1238,7 @@ static int parse_def_export( char *name, DLLSPEC *spec )
     return 1;
 
 error:
-    spec->entry_points.count--;
+    spec->nb_entry_points--;
     free( odp->name );
     return 0;
 }
