diff --git a/tools/winebuild/res16.c b/tools/winebuild/res16.c
index 94c1e96add2b036942b44ec4dc7e1ff9bb7e39bb..732af0151990ea339003c819f97b24823e77e7ee 100644
--- a/tools/winebuild/res16.c
+++ b/tools/winebuild/res16.c
@@ -56,9 +56,25 @@ struct res_type
     unsigned int             nb_names;     /* total number of names */
 };
 
-static struct res_type *add_type( struct array *types, struct resource *res )
+/* top level of the resource tree */
+struct res_tree
 {
-    struct res_type *type = ARRAY_ADD( types, struct res_type );
+    struct res_type *types;                /* types array */
+    unsigned int     nb_types;             /* total number of types */
+};
+
+
+static inline struct resource *add_resource( DLLSPEC *spec )
+{
+    spec->resources = xrealloc( spec->resources, (spec->nb_resources + 1) * sizeof(*spec->resources) );
+    return &spec->resources[spec->nb_resources++];
+}
+
+static struct res_type *add_type( struct res_tree *tree, struct resource *res )
+{
+    struct res_type *type;
+    tree->types = xrealloc( tree->types, (tree->nb_types + 1) * sizeof(*tree->types) );
+    type = &tree->types[tree->nb_types++];
     type->type        = &res->type;
     type->res         = res;
     type->nb_names    = 0;
@@ -86,7 +102,7 @@ static void get_string( struct string_id *str )
 /* load the next resource from the current file */
 static void load_next_resource( DLLSPEC *spec )
 {
-    struct resource *res = ARRAY_ADD( &spec->resources, struct resource );
+    struct resource *res = add_resource( spec );
 
     get_string( &res->type );
     get_string( &res->name );
@@ -130,22 +146,25 @@ static int cmp_res( const void *ptr1, const void *ptr2 )
 }
 
 /* build the 2-level (type,name) resource tree */
-static struct array build_resource_tree( DLLSPEC *spec )
+static struct res_tree *build_resource_tree( DLLSPEC *spec )
 {
-    unsigned int j, offset;
-    struct array types = empty_array;
+    unsigned int i, j, offset;
+    struct res_tree *tree;
     struct res_type *type = NULL;
     struct resource *res;
 
-    ARRAY_SORT( &spec->resources, struct resource, cmp_res );
+    qsort( spec->resources, spec->nb_resources, sizeof(*spec->resources), cmp_res );
 
     offset = 2;  /* alignment */
+    tree = xmalloc( sizeof(*tree) );
+    tree->types = NULL;
+    tree->nb_types = 0;
 
-    ARRAY_FOR_EACH( res, &spec->resources, struct resource )
+    for (i = 0; i < spec->nb_resources; i++)
     {
-        if (!type || cmp_string( &res->type, &res[-1].type ))  /* new type */
+        if (!i || cmp_string( &spec->resources[i].type, &spec->resources[i-1].type ))  /* new type */
         {
-            type = add_type( &types, res );
+            type = add_type( tree, &spec->resources[i] );
             offset += 8;
         }
         type->nb_names++;
@@ -153,7 +172,7 @@ static struct array build_resource_tree( DLLSPEC *spec )
     }
     offset += 2;  /* terminator */
 
-    ARRAY_FOR_EACH( type, &types, struct res_type )
+    for (i = 0, type = tree->types; i < tree->nb_types; i++, type++)
     {
         if (type->type->str)
         {
@@ -172,7 +191,14 @@ static struct array build_resource_tree( DLLSPEC *spec )
             else res->name_offset = res->name.id | 0x8000;
         }
     }
-    return types;
+    return tree;
+}
+
+/* free the resource tree */
+static void free_resource_tree( struct res_tree *tree )
+{
+    free( tree->types );
+    free( tree );
 }
 
 /* output a string preceded by its length */
@@ -194,11 +220,12 @@ static void output_bin_string( const char *str )
 /* output the resource data */
 void output_res16_data( DLLSPEC *spec )
 {
-    unsigned int i = 0;
+    const struct resource *res;
+    unsigned int i;
 
-    ARRAY_FOR_EACH( res, &spec->resources, const struct resource )
+    for (i = 0, res = spec->resources; i < spec->nb_resources; i++, res++)
     {
-        output( ".L__wine_spec_resource_%u:\n", i++ );
+        output( ".L__wine_spec_resource_%u:\n", i );
         dump_bytes( res->data, res->data_size );
     }
 }
@@ -206,23 +233,26 @@ void output_res16_data( DLLSPEC *spec )
 /* output the resource definitions */
 void output_res16_directory( DLLSPEC *spec )
 {
-    unsigned int j;
+    unsigned int i, j;
+    struct res_tree *tree;
+    const struct res_type *type;
     const struct resource *res;
-    struct array types = build_resource_tree( spec );
+
+    tree = build_resource_tree( spec );
 
     output( "\n.L__wine_spec_ne_rsrctab:\n" );
     output( "\t.short 0\n" );  /* alignment */
 
     /* type and name structures */
 
-    ARRAY_FOR_EACH( type, &types, const struct res_type )
+    for (i = 0, type = tree->types; i < tree->nb_types; i++, type++)
     {
         output( "\t.short 0x%04x,%u,0,0\n", type->name_offset, type->nb_names );
 
         for (j = 0, res = type->res; j < type->nb_names; j++, res++)
         {
             output( "\t.short .L__wine_spec_resource_%lu-.L__wine_spec_dos_header,%u\n",
-                    (unsigned long)(res - (struct resource *)spec->resources.data), res->data_size );
+                    (unsigned long)(res - spec->resources), res->data_size );
             output( "\t.short 0x%04x,0x%04x,0,0\n", res->memopt, res->name_offset );
         }
     }
@@ -230,33 +260,42 @@ void output_res16_directory( DLLSPEC *spec )
 
     /* name strings */
 
-    ARRAY_FOR_EACH( type, &types, const struct res_type )
+    for (i = 0, type = tree->types; i < tree->nb_types; i++, type++)
     {
         if (type->type->str) output_string( type->type->str );
         for (j = 0, res = type->res; j < type->nb_names; j++, res++)
             if (res->name.str) output_string( res->name.str );
     }
     output( "\t.byte 0\n" );  /* names terminator */
+
+    free_resource_tree( tree );
 }
 
 /* output the resource data in binary format */
 void output_bin_res16_data( DLLSPEC *spec )
 {
-    ARRAY_FOR_EACH( res, &spec->resources, const struct resource ) put_data( res->data, res->data_size );
+    const struct resource *res;
+    unsigned int i;
+
+    for (i = 0, res = spec->resources; i < spec->nb_resources; i++, res++)
+        put_data( res->data, res->data_size );
 }
 
 /* output the resource definitions in binary format */
 void output_bin_res16_directory( DLLSPEC *spec, unsigned int data_offset )
 {
-    unsigned int j;
+    unsigned int i, j;
+    struct res_tree *tree;
+    const struct res_type *type;
     const struct resource *res;
-    struct array types = build_resource_tree( spec );
+
+    tree = build_resource_tree( spec );
 
     put_word( 0 );  /* alignment */
 
     /* type and name structures */
 
-    ARRAY_FOR_EACH( type, &types, const struct res_type )
+    for (i = 0, type = tree->types; i < tree->nb_types; i++, type++)
     {
         put_word( type->name_offset );
         put_word( type->nb_names );
@@ -278,11 +317,13 @@ void output_bin_res16_directory( DLLSPEC *spec, unsigned int data_offset )
 
     /* name strings */
 
-    ARRAY_FOR_EACH( type, &types, const struct res_type )
+    for (i = 0, type = tree->types; i < tree->nb_types; i++, type++)
     {
         if (type->type->str) output_bin_string( type->type->str );
         for (j = 0, res = type->res; j < type->nb_names; j++, res++)
             if (res->name.str) output_bin_string( res->name.str );
     }
     put_byte( 0 );  /* names terminator */
+
+    free_resource_tree( tree );
 }
