diff --git a/tools/winedump/font.c b/tools/winedump/font.c
index ed4f4410e7e9b16dc81614f60dc7ccf90d351ca8..ca4065340c33ca69c5aa14d95451f35905ab946a 100644
--- a/tools/winedump/font.c
+++ b/tools/winedump/font.c
@@ -31,7 +31,7 @@
 
 #include "winedump.h"
 
-#pragma pack(push,1)
+#include <pshpack1.h>
 typedef struct
 {
     INT16 dfType;
@@ -77,7 +77,7 @@ typedef struct
     char dfCopyright[60];	/* Copyright notice */
     FONTINFO16 fi;		/* FONTINFO structure */
 } WINFNT;
-#pragma pack(pop)
+#include <poppack.h>
 
 /* FIXME: recognize and dump also NE/PE wrapped fonts */
 
