diff --git a/tools/winedump/winedump.h b/tools/winedump/winedump.h
index 61a3e55e67d38085ddff11ed7b4d618fcb875e0d..d15eacf2971c2d72ea9115ebc26127a3f3548ba5 100644
--- a/tools/winedump/winedump.h
+++ b/tools/winedump/winedump.h
@@ -228,8 +228,7 @@ void            dump_data_offset( const unsigned char *ptr, unsigned int size, u
 void            dump_data( const unsigned char *ptr, unsigned int size, const char *prefix );
 const char*	get_time_str( unsigned long );
 unsigned int    strlenW( const unsigned short *str );
-int             dump_strA( const char *str, size_t len );
-int             dump_strW( const WCHAR *str, size_t len );
+void            dump_unicode_str( const unsigned short *str, int len );
 const char*     get_hexint64_str( DWORD64 l );
 const char*     get_uint64_str( DWORD64 l );
 const char*     get_guid_str(const GUID* guid);
@@ -261,8 +260,8 @@ enum FileSig    get_kind_emfspool(void);
 void            emfspool_dump(void);
 enum FileSig    get_kind_mf(void);
 void            mf_dump(void);
-enum FileSig    get_kind_pdb(int fd);
-void            pdb_dump(int fd);
+enum FileSig    get_kind_pdb(void);
+void            pdb_dump(void);
 enum FileSig    get_kind_fnt(void);
 void            fnt_dump( void );
 enum FileSig    get_kind_tlb(void);
