diff --git a/tools/winemaker/winemaker b/tools/winemaker/winemaker
index d95b3096e51cb5c380b625e2b414dd080f1b508e..88ae78ce48702d0e1416566b033e1f3a0ba18f86 100755
--- a/tools/winemaker/winemaker
+++ b/tools/winemaker/winemaker
@@ -877,7 +877,7 @@ sub source_scan_project_file($$$)
             require XML::LibXML;
         };
         if ($@) {
-            die "Error: You need the libxml package (deb: libxml-libxml-perl, rpm: perl-libxml-perl, FreeBSD: p5-XML-LibXML)";
+            die "Error: You need the libxml package (deb: libxml-libxml-perl, rpm: perl-libxml-perl)";
         }
 
         my $xmlparser = XML::LibXML->new();
