diff --git a/tools/wmc/utils.c b/tools/wmc/utils.c
index 9bd961ead436722f0025a5102704b97a2c4a693a..6d8aa7572c320a61ae73dc40f33a53389470fbf8 100644
--- a/tools/wmc/utils.c
+++ b/tools/wmc/utils.c
@@ -88,6 +88,17 @@ void internal_error(const char *file, int line, const char *s, ...)
 	exit(3);
 }
 
+void fatal_perror( const char *msg, ... )
+{
+        va_list valist;
+        va_start( valist, msg );
+	fprintf(stderr, "Error: ");
+        vfprintf( stderr, msg, valist );
+        perror( " " );
+        va_end( valist );
+        exit(2);
+}
+
 void error(const char *s, ...)
 {
 	va_list ap;
