diff --git a/tools/wrc/newstruc.c b/tools/wrc/newstruc.c
index 6aa375139891a3df542571581e057578930bb267..9602cdfbf92f642d844dc189436b6f2f9af65d49 100644
--- a/tools/wrc/newstruc.c
+++ b/tools/wrc/newstruc.c
@@ -35,7 +35,7 @@
 
 #include "wingdi.h"	/* for BITMAPINFOHEADER */
 
-#pragma pack(push,2)
+#include <pshpack2.h>
 typedef struct
 {
     unsigned int   biSize;
@@ -44,7 +44,7 @@ typedef struct
     unsigned short biPlanes;
     unsigned short biBitCount;
 } BITMAPOS2HEADER;
-#pragma pack(pop)
+#include <poppack.h>
 
 /* New instances for all types of structures */
 /* Very inefficient (in size), but very functional :-]
@@ -156,6 +156,13 @@ lvc_t *new_lvc(void)
     return ret;
 }
 
+res_count_t *new_res_count(void)
+{
+    res_count_t *ret = xmalloc( sizeof(*ret) );
+    memset( ret, 0, sizeof(*ret) );
+    return ret;
+}
+
 string_t *new_string(void)
 {
     string_t *ret = xmalloc( sizeof(*ret) );
