diff --git a/tools/wrc/newstruc.h b/tools/wrc/newstruc.h
index ad3e56f43f3ee4527e0ffbe4ca8b7dbe62c0ce55..7ed386916abff490a3337f2f7a7009b254a85406 100644
--- a/tools/wrc/newstruc.h
+++ b/tools/wrc/newstruc.h
@@ -38,6 +38,7 @@ accelerator_t *new_accelerator(void);
 event_t *new_event(void);
 raw_data_t *new_raw_data(void);
 lvc_t *new_lvc(void);
+res_count_t *new_res_count(void);
 string_t *new_string(void);
 toolbar_item_t *new_toolbar_item(void);
 ani_any_t *new_ani_any(void);
