diff --git a/tools/wrc/utils.c b/tools/wrc/utils.c
index f0744638271dfd38fa60bb1b903a5c0450d5461d..b96547ae8ac6de09d26ef9f6cef197c78e4a06b3 100644
--- a/tools/wrc/utils.c
+++ b/tools/wrc/utils.c
@@ -86,6 +86,17 @@ int parser_warning(const char *s, ...)
 	return 0;
 }
 
+void fatal_perror( const char *msg, ... )
+{
+        va_list valist;
+        va_start( valist, msg );
+	fprintf(stderr, "Error: ");
+        vfprintf( stderr, msg, valist );
+        perror( " " );
+        va_end( valist );
+        exit(2);
+}
+
 void error(const char *s, ...)
 {
 	va_list ap;
