diff --git a/tools/wrc/utils.h b/tools/wrc/utils.h
index 62c0a41b253842f0e51dc91870b045a5675fc924..9b784919d667ffd73065b534d52934ce3600db18 100644
--- a/tools/wrc/utils.h
+++ b/tools/wrc/utils.h
@@ -27,6 +27,7 @@ int compare_striA( const char *str1, const char *str2 );
 int compare_striW( const WCHAR *str1, const WCHAR *str2 );
 int parser_error(const char *s, ...) __attribute__((format (printf, 1, 2)));
 int parser_warning(const char *s, ...) __attribute__((format (printf, 1, 2)));
+void fatal_perror( const char *msg, ... ) __attribute__((format (printf, 1, 2), noreturn));
 void error(const char *s, ...) __attribute__((format (printf, 1, 2), noreturn));
 void warning(const char *s, ...) __attribute__((format (printf, 1, 2)));
 void chat(const char *s, ...) __attribute__((format (printf, 1, 2)));
