diff --git a/tools/wrc/wrc.c b/tools/wrc/wrc.c
index 4e27d3ba5b4bc72b8f7eadc3d5068c17da27c8b9..15b63509f74dbddf35e32701db90f749330634b1 100644
--- a/tools/wrc/wrc.c
+++ b/tools/wrc/wrc.c
@@ -436,9 +436,9 @@ int main(int argc,char *argv[])
 
 	atexit(cleanup_files);
 
-        STRARRAY_FOR_EACH( file, &input_files )
+        for (i = 0; i < input_files.count; i++)
         {
-            input_name = file;
+            input_name = input_files.str[i];
             if (load_file( input_name, output_name )) exit(1);
         }
 	/* stdin special case. NULL means "stdin" for wpp. */
