diff --git a/tools/wrc/wrctypes.h b/tools/wrc/wrctypes.h
index 8c4d5a26f00a120a3ccb1288839eaa7d1bc1ace2..b453e2fdb08b4b5a68b81972c026dc5fbd054126 100644
--- a/tools/wrc/wrctypes.h
+++ b/tools/wrc/wrctypes.h
@@ -576,6 +576,22 @@ typedef struct resource {
 	unsigned int	memopt;
 } resource_t;
 
+/* Resource count */
+typedef struct res32_count {
+	int			count;
+	resource_t		**rsc;
+} res32_count_t;
+
+typedef struct res_count {
+	name_id_t		type;
+	int			count;		/* win16 mode */
+	resource_t		**rscarray;
+	int			count32;
+	res32_count_t		*rsc32array;	/* win32 mode */
+	int			n_id_entries;
+	int			n_name_entries;
+} res_count_t;
+
 typedef struct style_pair {
     	style_t			*style;
 	style_t			*exstyle;
