diff --git a/server/process.c b/server/process.c
--- a/server/process.c
+++ b/server/process.c
@@ -1140,7 +1140,8 @@ int set_process_debug_flag( struct process *process, int flag )
     char data = (flag != 0);
     client_ptr_t peb32 = 0;
 
-    if (is_wow64_process( process )) peb32 = process->peb + 0x1000;
+    if (!is_machine_64bit( process->machine ) && is_machine_64bit( native_machine ))
+        peb32 = process->peb + 0x1000;
 
     /* BeingDebugged flag is the byte at offset 2 in the PEB */
     if (peb32 && !write_process_memory( process, peb32 + 2, 1, &data, NULL )) return 0;
