diff --git a/server/process.c b/server/process.c
--- a/server/process.c
+++ b/server/process.c
@@ -1792,12 +1792,12 @@ DECL_HANDLER(read_process_memory)
 DECL_HANDLER(write_process_memory)
 {
     struct process *process;
+    reply->written = 0;
 
     if ((process = get_process_from_handle( req->handle, PROCESS_VM_WRITE )))
     {
         data_size_t len = get_req_data_size();
-        reply->written = 0;
         if (len) write_process_memory( process, req->addr, len, get_req_data(), &reply->written );
         release_object( process );
     }
 }
