diff --git a/server/thread.c b/server/thread.c
--- a/server/thread.c
+++ b/server/thread.c
@@ -2304,7 +2304,7 @@ DECL_HANDLER(set_thread_context)
         unsigned int flags = system_flags & contexts[CTX_NATIVE].flags;
 
         if (thread != current) stop_thread( thread );
-        else if (flags) set_thread_context( thread, &contexts[CTX_NATIVE], flags );
+        if (flags) set_thread_context( thread, &contexts[CTX_NATIVE], flags );
 
         if (thread->context && !get_error())
         {
